--- a/net/minecraft/core/dispenser/FlintAndSteelDispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/FlintAndSteelDispenseItemBehavior.java
@@ -3,17 +_,15 @@
 import java.util.List;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.core.Holder;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.monster.cubemob.SulfurCube;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.BaseFireBlock;
-import net.minecraft.world.level.block.CampfireBlock;
-import net.minecraft.world.level.block.CandleBlock;
-import net.minecraft.world.level.block.CandleCakeBlock;
 import net.minecraft.world.level.block.DispenserBlock;
 import net.minecraft.world.level.block.TntBlock;
 import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 
@@ -24,17 +_,30 @@
         this.setSuccess(true);
         Direction facing = source.state().getValue(DispenserBlock.FACING);
         BlockPos targetPos = source.pos().relative(facing);
+        // Paper start - Call BlockDispenseEvent
+        ItemStack result = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockDispenseEvent(source, targetPos, dispensed, this);
+        if (result != null) {
+            this.setSuccess(false);
+            return result;
+        }
+        // Paper end - Call BlockDispenseEvent
         BlockState target = level.getBlockState(targetPos);
         if (!tryIgniteExplosiveEntities(level, targetPos)) {
             if (BaseFireBlock.canBePlacedAt(level, targetPos, facing)) {
-                level.setBlockAndUpdate(targetPos, BaseFireBlock.getState(level, targetPos));
-                level.gameEvent(null, GameEvent.BLOCK_PLACE, targetPos);
-            } else if (CampfireBlock.canLight(target) || CandleBlock.canLight(target) || CandleCakeBlock.canLight(target)) {
-                level.setBlockAndUpdate(targetPos, target.setValue(BlockStateProperties.LIT, true));
+                // CraftBukkit start - Ignition by dispensing flint and steel
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, targetPos, source.pos()).isCancelled()) {
+                    level.setBlockAndUpdate(targetPos, BaseFireBlock.getState(level, targetPos));
+                    level.gameEvent((Entity) null, (Holder) GameEvent.BLOCK_PLACE, targetPos);
+                }
+                // CraftBukkit end
+            } else if (target.getToolModifiedState(new net.minecraft.world.item.context.UseOnContext(level, null, net.minecraft.world.InteractionHand.MAIN_HAND, dispensed, new net.minecraft.world.phys.BlockHitResult(net.minecraft.world.phys.Vec3.atCenterOf(targetPos), facing.getOpposite(), targetPos, false)), net.neoforged.neoforge.common.ItemAbilities.FIRESTARTER_LIGHT, false) instanceof BlockState blockstate2) {
+                level.setBlockAndUpdate(targetPos, blockstate2);
                 level.gameEvent(null, GameEvent.BLOCK_CHANGE, targetPos);
-            } else if (target.getBlock() instanceof TntBlock) {
-                if (TntBlock.prime(level, targetPos, null, dispensed)) {
-                    level.removeBlock(targetPos, false);
+            } else if (target.isFlammable(level, targetPos, source.state().getValue(DispenserBlock.FACING).getOpposite())) {
+                if (target.onCaughtFire(level, targetPos, source.state().getValue(DispenserBlock.FACING).getOpposite(), null)) {
+                    if (target.getBlock() instanceof TntBlock && org.bukkit.craftbukkit.event.CraftEventFactory.callTNTPrimeEvent(level, targetPos, org.bukkit.event.block.TNTPrimeEvent.PrimeCause.DISPENSER, null, source.pos())) { // CraftBukkit - TNTPrimeEvent
+                        level.removeBlock(targetPos, false);
+                    }
                 } else {
                     this.setSuccess(false);
                 }
