--- a/net/minecraft/world/level/levelgen/synth/SimplexNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/SimplexNoise.java
@@ -1,27 +_,59 @@
 package net.minecraft.world.level.levelgen.synth;
 
-import net.minecraft.util.Interval;
 import net.minecraft.util.Mth;
 import net.minecraft.util.RandomSource;
 
-public class SimplexNoise extends GradientNoise {
-    public static final Interval RANGE = Interval.ofSymmetric(2.0F);
-    public static final double STANDARD_DEVIATION = 0.42544;
+public class SimplexNoise {
+    protected static final int[][] GRADIENT = new int[][]{
+        {1, 1, 0},
+        {-1, 1, 0},
+        {1, -1, 0},
+        {-1, -1, 0},
+        {1, 0, 1},
+        {-1, 0, 1},
+        {1, 0, -1},
+        {-1, 0, -1},
+        {0, 1, 1},
+        {0, -1, 1},
+        {0, 1, -1},
+        {0, -1, -1},
+        {1, 1, 0},
+        {0, -1, 1},
+        {-1, 1, 0},
+        {0, -1, -1}
+    };
     private static final double SQRT_3 = Math.sqrt(3.0);
     private static final double F2 = 0.5 * (SQRT_3 - 1.0);
     private static final double G2 = (3.0 - SQRT_3) / 6.0;
+    private final int[] p = new int[512];
+    public final double xo;
+    public final double yo;
+    public final double zo;
 
     public SimplexNoise(RandomSource random) {
-        super(random);
-    }
-
-    public SimplexNoise(RandomSource random, boolean discardNoiseOffset) {
-        super(random, discardNoiseOffset ? 0.0 : 256.0);
-    }
-
-    @Override
-    public Interval range() {
-        return RANGE;
+        this.xo = random.nextDouble() * 256.0;
+        this.yo = random.nextDouble() * 256.0;
+        this.zo = random.nextDouble() * 256.0;
+        int i = 0;
+
+        while (i < 256) {
+            this.p[i] = i++;
+        }
+
+        for (int ix = 0; ix < 256; ix++) {
+            int offset = random.nextInt(256 - ix);
+            int tmp = this.p[ix];
+            this.p[ix] = this.p[offset + ix];
+            this.p[offset + ix] = tmp;
+        }
+    }
+
+    private int p(int x) {
+        return this.p[x & 0xFF];
+    }
+
+    protected static double dot(int[] g, double x, double y, double z) {
+        return g[0] * x + g[1] * y + g[2] * z;
     }
 
     private double getCornerNoise3D(int index, double x, double y, double z, double base) {
@@ -31,16 +_,13 @@
             n0 = 0.0;
         } else {
             t0 *= t0;
-            n0 = t0 * t0 * GRADIENT[index].dot(x, y, z);
+            n0 = t0 * t0 * dot(GRADIENT[index], x, y, z);
         }
 
         return n0;
     }
 
-    @Override
-    public float get(double _xin, double _yin) {
-        double xin = _xin + this.offsetX;
-        double yin = _yin + this.offsetY;
+    public double getValue(double xin, double yin) {
         double s = (xin + yin) * F2;
         int i = Mth.floor(xin + s);
         int j = Mth.floor(yin + s);
@@ -65,20 +_,16 @@
         double y2 = y0 - 1.0 + 2.0 * G2;
         int ii = i & 0xFF;
         int jj = j & 0xFF;
-        int gi0 = this.permute(ii + this.permute(jj)) % 12;
-        int gi1 = this.permute(ii + i1 + this.permute(jj + j1)) % 12;
-        int gi2 = this.permute(ii + 1 + this.permute(jj + 1)) % 12;
+        int gi0 = this.p(ii + this.p(jj)) % 12;
+        int gi1 = this.p(ii + i1 + this.p(jj + j1)) % 12;
+        int gi2 = this.p(ii + 1 + this.p(jj + 1)) % 12;
         double n0 = this.getCornerNoise3D(gi0, x0, y0, 0.0, 0.5);
         double n1 = this.getCornerNoise3D(gi1, x1, y1, 0.0, 0.5);
         double n2 = this.getCornerNoise3D(gi2, x2, y2, 0.0, 0.5);
-        return (float)(70.0 * (n0 + n1 + n2));
+        return 70.0 * (n0 + n1 + n2);
     }
 
-    @Override
-    public float get(double _xin, double _yin, double _zin) {
-        double xin = _xin + this.offsetX;
-        double yin = _yin + this.offsetY;
-        double zin = _zin + this.offsetZ;
+    public double getValue(double xin, double yin, double zin) {
         double F3 = 0.3333333333333333;
         double s = (xin + yin + zin) * 0.3333333333333333;
         int i = Mth.floor(xin + s);
@@ -156,14 +_,14 @@
         int ii = i & 0xFF;
         int jj = j & 0xFF;
         int kk = k & 0xFF;
-        int gi0 = this.permute(ii + this.permute(jj + this.permute(kk))) % 12;
-        int gi1 = this.permute(ii + i1 + this.permute(jj + j1 + this.permute(kk + k1))) % 12;
-        int gi2 = this.permute(ii + i2 + this.permute(jj + j2 + this.permute(kk + k2))) % 12;
-        int gi3 = this.permute(ii + 1 + this.permute(jj + 1 + this.permute(kk + 1))) % 12;
+        int gi0 = this.p(ii + this.p(jj + this.p(kk))) % 12;
+        int gi1 = this.p(ii + i1 + this.p(jj + j1 + this.p(kk + k1))) % 12;
+        int gi2 = this.p(ii + i2 + this.p(jj + j2 + this.p(kk + k2))) % 12;
+        int gi3 = this.p(ii + 1 + this.p(jj + 1 + this.p(kk + 1))) % 12;
         double n0 = this.getCornerNoise3D(gi0, x0, y0, z0, 0.6);
         double n1 = this.getCornerNoise3D(gi1, x1, y1, z1, 0.6);
         double n2 = this.getCornerNoise3D(gi2, x2, y2, z2, 0.6);
         double n3 = this.getCornerNoise3D(gi3, x3, y3, z3, 0.6);
-        return (float)(32.0 * (n0 + n1 + n2 + n3));
+        return 32.0 * (n0 + n1 + n2 + n3);
     }
 }
