--- a/net/minecraft/server/packs/PackLocationInfo.java
+++ b/net/minecraft/server/packs/PackLocationInfo.java
@@ -6,7 +_,6 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentUtils;
 import net.minecraft.network.chat.HoverEvent;
-import net.minecraft.network.chat.Style;
 import net.minecraft.server.packs.repository.KnownPack;
 import net.minecraft.server.packs.repository.PackSource;
 
