--- a/net/minecraft/server/packs/PathPackResources.java
+++ b/net/minecraft/server/packs/PathPackResources.java
@@ -105,6 +_,11 @@
         try (DirectoryStream<Path> directorystream = Files.newDirectoryStream(path)) {
             for (Path path1 : directorystream) {
                 String s = path1.getFileName().toString();
+                // Paper start - Improve logging and errors
+                if (!Files.isDirectory(path1)) {
+                    continue;
+                }
+                // Paper end - Improve logging and errors
                 if (ResourceLocation.isValidNamespace(s)) {
                     set.add(s);
                 } else {
