--- a/net/minecraft/server/packs/VanillaPackResources.java
+++ b/net/minecraft/server/packs/VanillaPackResources.java
@@ -1,13 +_,16 @@
 package net.minecraft.server.packs;
 
+import java.io.InputStream;
 import java.nio.file.Path;
 import java.util.List;
+import java.util.Optional;
 import java.util.function.Consumer;
 import java.util.stream.Stream;
 import net.minecraft.resources.Identifier;
 import net.minecraft.server.packs.repository.Pack;
-import net.minecraft.server.packs.resources.FallbackResourceManager;
-import net.minecraft.server.packs.resources.ResourceManager;
+import net.minecraft.server.packs.resources.IoSupplier;
+import net.minecraft.server.packs.resources.Resource;
+import net.minecraft.server.packs.resources.ResourceProvider;
 
 public class VanillaPackResources {
     private final FixedPathPackResources fullResources;
@@ -40,9 +_,8 @@
         };
     }
 
-    public ResourceManager asResourceManager() {
-        FallbackResourceManager vanillaOnlyResourceManager = new FallbackResourceManager(PackType.CLIENT_RESOURCES, "minecraft");
-        vanillaOnlyResourceManager.push(this.fullResources);
-        return vanillaOnlyResourceManager;
+    public ResourceProvider asProvider() {
+        return location -> Optional.ofNullable(this.fullResources.getResource(PackType.CLIENT_RESOURCES, location))
+            .map(s -> new Resource(this.fullResources, (IoSupplier<InputStream>)s));
     }
 }
