--- a/net/minecraft/world/entity/animal/bee/Bee.java
+++ b/net/minecraft/world/entity/animal/bee/Bee.java
@@ -5,6 +_,7 @@
 import java.util.Comparator;
 import java.util.EnumSet;
 import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -53,6 +_,7 @@
 import net.minecraft.world.entity.ai.goal.FloatGoal;
 import net.minecraft.world.entity.ai.goal.FollowParentGoal;
 import net.minecraft.world.entity.ai.goal.Goal;
+import net.minecraft.world.entity.ai.goal.GoalSelector;
 import net.minecraft.world.entity.ai.goal.MeleeAttackGoal;
 import net.minecraft.world.entity.ai.goal.TemptGoal;
 import net.minecraft.world.entity.ai.goal.target.HurtByTargetGoal;
@@ -74,7 +_,6 @@
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
-import net.minecraft.world.level.block.BonemealSource;
 import net.minecraft.world.level.block.BonemealableBlock;
 import net.minecraft.world.level.block.CropBlock;
 import net.minecraft.world.level.block.DoublePlantBlock;
@@ -94,6 +_,12 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+import org.bukkit.event.entity.EntityTargetEvent;
+// CraftBukkit end
+
 public class Bee extends Animal implements NeutralMob {
     public static final float FLAP_DEGREES_PER_TICK = 120.32113F;
     public static final int TICKS_PER_FLAP = Mth.ceil(1.4959966F);
@@ -147,18 +_,104 @@
     private Bee.BeeGoToHiveGoal goToHiveGoal;
     private Bee.BeeGoToKnownFlowerGoal goToKnownFlowerGoal;
     private int underWaterTicks;
+    public net.kyori.adventure.util.TriState rollingOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - Rolling override
+
+    // Purpur start - Make entity breeding times configurable
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level().purpurConfig.beeBreedingTicks;
+    }
+    // Purpur end - Make entity breeding times configurable
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.beeAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
 
     public Bee(EntityType<? extends Bee> type, Level level) {
         super(type, level);
-        this.moveControl = new FlyingMoveControl<>(this, 20, true);
+        final org.purpurmc.purpur.controller.FlyingMoveControllerWASD flyingController = new org.purpurmc.purpur.controller.FlyingMoveControllerWASD(this, 0.25F, 1.0F, false); // Purpur - Ridables
+        // Paper start - Fix MC-167279
+        class BeeFlyingMoveControl extends FlyingMoveControl<Mob> {
+            public BeeFlyingMoveControl(final Mob mob, final int maxTurn, final boolean hoversInPlace) {
+                super(mob, maxTurn, hoversInPlace);
+            }
+
+            @Override
+            public void tick() {
+                // Purpur start - Ridables
+                if (mob.getRider() != null && mob.isControllable()) {
+                    flyingController.purpurTick(mob.getRider());
+                    return;
+                }
+                // Purpur end - Ridables
+                if (this.mob.getY() <= Bee.this.level().getMinY()) {
+                    this.mob.setNoGravity(false);
+                }
+                super.tick();
+            }
+
+            // Purpur start - Ridables
+            @Override
+            public boolean hasWanted() {
+                return mob.getRider() != null || !mob.isControllable() || super.hasWanted();
+            }
+            // Purpur end - Ridables
+        }
+        this.moveControl = new BeeFlyingMoveControl(this, 20, true);
+        // Paper end - Fix MC-167279
         this.lookControl = new Bee.BeeLookControl(this);
         this.setPathfindingMalus(PathType.FIRE, -1.0F);
-        this.setPathfindingMalus(PathType.WATER, -1.0F);
+        if (this.level().purpurConfig.beeCanInstantlyStartDrowning || isSensitiveToWater()) this.setPathfindingMalus(PathType.WATER, -1.0F); // Purpur - bee can instantly start drowning in water option // Purpur - Toggle for water sensitive mob damage
         this.setPathfindingMalus(PathType.WATER_BORDER, 16.0F);
         this.setPathfindingMalus(PathType.COCOA, -1.0F);
         this.setPathfindingMalus(PathType.FENCE, -1.0F);
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.beeRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.beeRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.beeControllable;
+    }
+
+    @Override
+    public double getMaxY() {
+        return level().purpurConfig.beeMaxY;
+    }
+
+    @Override
+    public void travel(Vec3 vec3) {
+        super.travel(vec3);
+        if (getRider() != null && this.isControllable() && !onGround) {
+            float speed = (float) getAttributeValue(Attributes.FLYING_SPEED) * 2;
+            setSpeed(speed);
+            Vec3 mot = getDeltaMovement();
+            move(net.minecraft.world.entity.MoverType.SELF, mot.multiply(speed, speed, speed));
+            setDeltaMovement(mot.scale(0.9D));
+        }
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.beeMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.beeScale);
+    }
+    // Purpur end - Configurable entity base attributes
+
     @Override
     protected void defineSynchedData(SynchedEntityData.Builder entityData) {
         super.defineSynchedData(entityData);
@@ -173,6 +_,7 @@
 
     @Override
     protected void registerGoals() {
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(0, new Bee.BeeAttackGoal(this, 1.4F, true));
         this.goalSelector.addGoal(1, new Bee.BeeEnterHiveGoal());
         this.goalSelector.addGoal(2, new BreedGoal(this, 1.0));
@@ -190,6 +_,7 @@
         this.goalSelector.addGoal(7, new Bee.BeeGrowCropGoal());
         this.goalSelector.addGoal(8, new Bee.BeeWanderGoal());
         this.goalSelector.addGoal(9, new FloatGoal(this));
+        this.targetSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur - Ridables
         this.targetSelector.addGoal(1, new Bee.BeeHurtByOtherGoal(this).setAlertOthers());
         this.targetSelector.addGoal(2, new Bee.BeeBecomeAngryTargetGoal(this));
         this.targetSelector.addGoal(3, new ResetUniversalAngerTargetGoal<>(this, true));
@@ -197,9 +_,24 @@
 
     @Override
     protected void addAdditionalSaveData(ValueOutput output) {
+        // CraftBukkit start - selectively save data
+        addAdditionalSaveData(output, true);
+    }
+
+    @Override
+    protected void addAdditionalSaveData(ValueOutput output, boolean includeAll) {
+        // CraftBukkit end
         super.addAdditionalSaveData(output);
-        output.storeNullable("hive_pos", BlockPos.CODEC, this.hivePos);
-        output.storeNullable("flower_pos", BlockPos.CODEC, this.savedFlowerPos);
+        // CraftBukkit start - selectively save hive
+        if (includeAll) {
+            output.storeNullable("hive_pos", BlockPos.CODEC, this.hivePos);
+        }
+        // CraftBukkit end
+        // CraftBukkit start - selectively save flower
+        if (includeAll) {
+            output.storeNullable("flower_pos", BlockPos.CODEC, this.savedFlowerPos);
+        }
+        // CraftBukkit end
         output.putBoolean("HasNectar", this.hasNectar());
         output.putBoolean("HasStung", this.hasStung());
         output.putInt("TicksSincePollination", this.ticksWithoutNectarSinceExitingHive);
@@ -237,7 +_,7 @@
                 }
 
                 if (poisonTime > 0) {
-                    livingTarget.addEffect(new MobEffectInstance(MobEffects.POISON, poisonTime * 20, 0), this);
+                    livingTarget.addEffect(new MobEffectInstance(MobEffects.POISON, poisonTime * 20, 0), this, EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
                 }
             }
 
@@ -333,13 +_,19 @@
         if (this.stayOutOfHiveCountdown <= 0 && !this.beePollinateGoal.isPollinating() && !this.hasStung() && this.getTarget() == null) {
             boolean wantsToEnterHive = this.hasNectar()
                 || this.isTiredOfLookingForNectar()
-                || this.level().environmentAttributes().getValue(EnvironmentAttributes.BEES_STAY_IN_HIVE, this.position());
+                || this.level().environmentAttributes().getValue(EnvironmentAttributes.BEES_STAY_IN_HIVE, this.position()) || isNightOrRaining(this.level()); // Purpur - Bee can work when raining or at night
             return wantsToEnterHive && !this.isHiveNearFire();
         } else {
             return false;
         }
     }
 
+    // Purpur start - Bee can work when raining or at night
+    public static boolean isNightOrRaining(Level level) {
+        return level.dimensionType().hasSkyLight() && (level.isDarkOutside() && !level.purpurConfig.beeCanWorkAtNight || level.isRaining() && !level.purpurConfig.beeCanWorkInRain); // Purpur - Bee can work when raining or at night
+    }
+    // Purpur end - Bee can work when raining or at night
+
     public void setStayOutOfHiveCountdown(int ticks) {
         this.stayOutOfHiveCountdown = ticks;
     }
@@ -360,7 +_,7 @@
     @Override
     protected void customServerAiStep(ServerLevel level) {
         boolean hasStung = this.hasStung();
-        if (this.isInWater()) {
+        if (this.level().purpurConfig.beeCanInstantlyStartDrowning && this.isInWater()) { // Purpur - bee can instantly start drowning in water option
             this.underWaterTicks++;
         } else {
             this.underWaterTicks = 0;
@@ -370,6 +_,7 @@
             this.hurtServer(level, this.damageSources().drown(), 1.0F);
         }
 
+        if (hasStung && !this.level().purpurConfig.beeDiesAfterSting) setHasStung(false); else // Purpur - Stop bees from dying after stinging
         if (hasStung) {
             this.timeSinceSting++;
             if (this.timeSinceSting % 5 == 0 && this.random.nextInt(Mth.clamp(1200 - this.timeSinceSting, 1, 1200)) == 0) {
@@ -472,7 +_,11 @@
         if (this.hivePos == null) {
             return null;
         } else {
-            return this.isTooFarAway(this.hivePos) ? null : this.level().getBlockEntity(this.hivePos, BlockEntityTypes.BEEHIVE).orElse(null);
+            // Paper start - move over logic to accommodate isTooFarAway with chunk load check
+            return this.isTooFarAway(this.hivePos) || this.level().getChunkIfLoadedImmediately(this.hivePos.getX() >> 4, this.hivePos.getZ() >> 4) == null
+                ? null
+                : this.level().getBlockEntity(this.hivePos) instanceof BeehiveBlockEntity hive ? hive : null;
+            // Paper end
         }
     }
 
@@ -504,7 +_,8 @@
         return this.getFlag(2);
     }
 
-    public void setRolling(boolean rolling) {
+    public void setRolling(boolean rolling) { // Paper - Rolling override
+        rolling = this.rollingOverride.toBooleanOrElse(rolling); // Paper - Rolling override
         this.setFlag(2, rolling);
     }
 
@@ -561,7 +_,7 @@
             if (effect != null) {
                 this.usePlayerItem(player, hand, heldItem);
                 if (!this.level().isClientSide()) {
-                    this.addEffect(effect);
+                    this.addEffect(effect, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.FOOD); // Paper - Add missing effect cause
                 }
 
                 return InteractionResult.SUCCESS;
@@ -631,18 +_,34 @@
     public boolean hurtServer(ServerLevel level, DamageSource source, float damage) {
         if (this.isInvulnerableTo(level, source)) {
             return false;
+        } else {
+            // CraftBukkit start - Only stop pollinating if entity was damaged
+            boolean result = super.hurtServer(level, source, damage);
+            if (!result) {
+                return result;
+            }
+            // CraftBukkit end
+            this.beePollinateGoal.stopPollinating();
+            return result; // CraftBukkit
         }
-
-        this.beePollinateGoal.stopPollinating();
-        return super.hurtServer(level, source, damage);
     }
 
     @Override
+    @Deprecated // FORGE: use jumpInFluid instead
     protected void jumpInLiquid(TagKey<Fluid> type) {
+        this.jumpInLiquidInternal();
+    }
+
+    private void jumpInLiquidInternal() {
         this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.01, 0.0));
     }
 
     @Override
+    public void jumpInFluid(net.neoforged.neoforge.fluids.FluidType type) {
+        this.jumpInLiquidInternal();
+    }
+
+    @Override
     public Vec3 getLeashOffset() {
         return new Vec3(0.0, 0.5F * this.getEyeHeight(), this.getBbWidth() * 0.2F);
     }
@@ -978,13 +_,13 @@
                             }
                         } else if (belowState.is(Blocks.CAVE_VINES) || belowState.is(Blocks.CAVE_VINES_PLANT)) {
                             BonemealableBlock bonemealableBlock = (BonemealableBlock)belowState.getBlock();
-                            if (bonemealableBlock.isValidBonemealTarget(Bee.this.level(), belowPos, belowState, BonemealSource.MOB)) {
-                                bonemealableBlock.performBonemeal((ServerLevel)Bee.this.level(), Bee.this.random, belowPos, belowState, BonemealSource.MOB);
+                            if (bonemealableBlock.isValidBonemealTarget(Bee.this.level(), belowPos, belowState)) {
+                                bonemealableBlock.performBonemeal((ServerLevel)Bee.this.level(), Bee.this.random, belowPos, belowState);
                                 growState = Bee.this.level().getBlockState(belowPos);
                             }
                         }
 
-                        if (growState != null) {
+                        if (growState != null && CraftEventFactory.callEntityChangeBlockEvent(Bee.this, belowPos, growState)) { // CraftBukkit
                             Bee.this.level().levelEvent(2011, belowPos, 15);
                             Bee.this.level().setBlockAndUpdate(belowPos, growState);
                             Bee.this.incrementNumCropsGrownSincePollination();
@@ -1008,7 +_,7 @@
         @Override
         protected void alertOther(Mob other, LivingEntity hurtByMob) {
             if (other instanceof Bee && this.mob.hasLineOfSight(hurtByMob)) {
-                other.setTarget(hurtByMob);
+                other.setTarget(hurtByMob, EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY); // CraftBukkit - reason
             }
         }
     }
@@ -1104,6 +_,7 @@
                     Bee.this.savedFlowerPos = nearbyPos.get();
                     Bee.this.navigation
                         .moveTo(Bee.this.savedFlowerPos.getX() + 0.5, Bee.this.savedFlowerPos.getY() + 0.5, Bee.this.savedFlowerPos.getZ() + 0.5, 1.2F);
+                    new org.purpurmc.purpur.event.entity.BeeFoundFlowerEvent((org.bukkit.entity.Bee) Bee.this.getBukkitEntity(), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(Bee.this.savedFlowerPos, Bee.this.level())).callEvent(); // Purpur - Bee API
                     return true;
                 } else {
                     Bee.this.remainingCooldownBeforeLocatingNewFlower = Mth.nextInt(Bee.this.random, 20, 60);
@@ -1155,6 +_,7 @@
             this.pollinating = false;
             Bee.this.navigation.stop();
             Bee.this.remainingCooldownBeforeLocatingNewFlower = 200;
+            new org.purpurmc.purpur.event.entity.BeeStopPollinatingEvent((org.bukkit.entity.Bee) Bee.this.getBukkitEntity(), Bee.this.savedFlowerPos == null ? null : org.bukkit.craftbukkit.util.CraftLocation.toBukkit(Bee.this.savedFlowerPos, Bee.this.level()), Bee.this.hasNectar()).callEvent(); // Purpur - Bee API
         }
 
         @Override
@@ -1170,7 +_,7 @@
                     Bee.this.dropFlower();
                     this.pollinating = false;
                     Bee.this.remainingCooldownBeforeLocatingNewFlower = 200;
-                } else {
+                } else if (Bee.this.savedFlowerPos != null) { // Paper - add null check since API can manipulate this
                     Vec3 flowerPos = Vec3.atBottomCenterOf(Bee.this.savedFlowerPos).add(0.0, 0.6F, 0.0);
                     if (flowerPos.distanceTo(Bee.this.position()) > 1.0) {
                         this.hoverPos = flowerPos;
@@ -1201,6 +_,7 @@
                                 this.setWantedPos();
                             }
 
+                            if (this.successfulPollinatingTicks == 0) new org.purpurmc.purpur.event.entity.BeeStartedPollinatingEvent((org.bukkit.entity.Bee) Bee.this.getBukkitEntity(), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(Bee.this.savedFlowerPos, Bee.this.level())).callEvent(); // Purpur - Bee API
                             this.successfulPollinatingTicks++;
                             if (Bee.this.random.nextFloat() < 0.05F && this.successfulPollinatingTicks > this.lastSoundPlayedTick + 60) {
                                 this.lastSoundPlayedTick = this.successfulPollinatingTicks;
@@ -1221,27 +_,25 @@
         }
 
         private Optional<BlockPos> findNearbyFlower() {
+            Iterable<BlockPos> closestNearbyFlowers = BlockPos.withinManhattan(Bee.this.blockPosition(), 5, 5, 5);
             Long2LongOpenHashMap tempCache = new Long2LongOpenHashMap();
-            Optional<BlockPos> result = Bee.this.level()
-                .findBlocksInBoxByManhattanDistance(Bee.this.blockPosition(), 5)
-                .filterState(Bee::attractsBees)
-                .findFirst((pos, var3) -> {
-                    long unreachableUntilTime = this.unreachableFlowerCache.getOrDefault(pos.asLong(), Long.MIN_VALUE);
-                    if (Bee.this.level().getGameTime() < unreachableUntilTime) {
-                        tempCache.put(pos.asLong(), unreachableUntilTime);
-                        return false;
-                    }
 
+            for (BlockPos pos : closestNearbyFlowers) {
+                long unreachableUntilTime = this.unreachableFlowerCache.getOrDefault(pos.asLong(), Long.MIN_VALUE);
+                if (Bee.this.level().getGameTime() < unreachableUntilTime) {
+                    tempCache.put(pos.asLong(), unreachableUntilTime);
+                } else if (Bee.attractsBees(Bee.this.level().getBlockState(pos))) {
                     Path path = Bee.this.navigation.createPath(pos, 1);
                     if (path != null && path.canReach()) {
-                        return true;
+                        return Optional.of(pos);
                     }
 
                     tempCache.put(pos.asLong(), Bee.this.level().getGameTime() + 600L);
-                    return false;
-                });
+                }
+            }
+
             this.unreachableFlowerCache = tempCache;
-            return result;
+            return Optional.empty();
         }
     }
 
