--- a/net/minecraft/commands/execution/tasks/BuildContexts.java
+++ b/net/minecraft/commands/execution/tasks/BuildContexts.java
@@ -2,6 +_,7 @@
 
 import com.google.common.annotations.VisibleForTesting;
 import com.mojang.brigadier.RedirectModifier;
+import com.mojang.brigadier.Command;
 import com.mojang.brigadier.context.CommandContext;
 import com.mojang.brigadier.context.ContextChain;
 import com.mojang.brigadier.context.ContextChain.Stage;
@@ -95,7 +_,7 @@
             }
         } else {
             CommandContext<T> commandcontext1 = contextchain.getTopContext();
-            com.mojang.brigadier.Command<T> command = commandcontext1.getCommand();
+            Command<T> command = commandcontext1.getCommand();
             if (command instanceof CustomCommandExecutor<?>) {
                 CustomCommandExecutor<T> customcommandexecutor = (CustomCommandExecutor<T>) command;
                 ExecutionControl<T> executioncontrol = ExecutionControl.create(p_305977_, p_309603_);
