--- a/net/minecraft/world/level/levelgen/feature/rootplacers/AboveRootPlacement.java
+++ b/net/minecraft/world/level/levelgen/feature/rootplacers/AboveRootPlacement.java
@@ -2,10 +_,9 @@
 
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import net.minecraft.core.Holder;
 import net.minecraft.world.level.levelgen.feature.stateproviders.BlockStateProvider;
 
-public record AboveRootPlacement(Holder<BlockStateProvider> aboveRootProvider, float aboveRootPlacementChance) {
+public record AboveRootPlacement(BlockStateProvider aboveRootProvider, float aboveRootPlacementChance) {
     public static final Codec<AboveRootPlacement> CODEC = RecordCodecBuilder.create(
         i -> i.group(
                 BlockStateProvider.CODEC.fieldOf("above_root_provider").forGetter(c -> c.aboveRootProvider),
