--- a/net/minecraft/world/level/levelgen/feature/rootplacers/MangroveRootPlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/rootplacers/MangroveRootPlacer.java
@@ -8,7 +_,6 @@
 import java.util.function.BiConsumer;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
-import net.minecraft.core.Holder;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.valueproviders.IntProvider;
 import net.minecraft.world.level.LevelSimulatedReader;
@@ -28,10 +_,7 @@
     private final MangroveRootPlacement mangroveRootPlacement;
 
     public MangroveRootPlacer(
-        IntProvider trunkOffsetY,
-        Holder<BlockStateProvider> rootProvider,
-        Optional<AboveRootPlacement> aboveRootPlacement,
-        MangroveRootPlacement mangroveRootPlacement
+        IntProvider trunkOffsetY, BlockStateProvider rootProvider, Optional<AboveRootPlacement> aboveRootPlacement, MangroveRootPlacement mangroveRootPlacement
     ) {
         super(trunkOffsetY, rootProvider, aboveRootPlacement);
         this.mangroveRootPlacement = mangroveRootPlacement;
@@ -117,7 +_,7 @@
     @Override
     protected void placeRoot(WorldGenLevel level, BiConsumer<BlockPos, BlockState> rootSetter, RandomSource random, BlockPos pos, TreeFeature tree) {
         if (level.isStateAtPosition(pos, s -> s.is(this.mangroveRootPlacement.muddyRootsIn()))) {
-            BlockState muddyRoots = this.mangroveRootPlacement.muddyRootsProvider().value().getState(level, random, pos);
+            BlockState muddyRoots = this.mangroveRootPlacement.muddyRootsProvider().getState(level, random, pos);
             rootSetter.accept(pos, this.getPotentiallyWaterloggedState(level, pos, muddyRoots));
         } else {
             super.placeRoot(level, rootSetter, random, pos, tree);
