--- a/net/minecraft/world/level/levelgen/feature/rootplacers/RootPlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/rootplacers/RootPlacer.java
@@ -7,7 +_,6 @@
 import java.util.Optional;
 import java.util.function.BiConsumer;
 import net.minecraft.core.BlockPos;
-import net.minecraft.core.Holder;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.RandomSource;
@@ -24,12 +_,10 @@
 public abstract class RootPlacer {
     public static final Codec<RootPlacer> CODEC = BuiltInRegistries.ROOT_PLACER_TYPE.byNameCodec().dispatch(RootPlacer::type, RootPlacerType::codec);
     protected final IntProvider trunkOffsetY;
-    protected final Holder<BlockStateProvider> rootProvider;
+    protected final BlockStateProvider rootProvider;
     protected final Optional<AboveRootPlacement> aboveRootPlacement;
 
-    protected static <P extends RootPlacer> P3<Mu<P>, IntProvider, Holder<BlockStateProvider>, Optional<AboveRootPlacement>> rootPlacerParts(
-        Instance<P> instance
-    ) {
+    protected static <P extends RootPlacer> P3<Mu<P>, IntProvider, BlockStateProvider, Optional<AboveRootPlacement>> rootPlacerParts(Instance<P> instance) {
         return instance.group(
             IntProviders.CODEC.fieldOf("trunk_offset_y").forGetter(c -> c.trunkOffsetY),
             BlockStateProvider.CODEC.fieldOf("root_provider").forGetter(c -> c.rootProvider),
@@ -37,7 +_,7 @@
         );
     }
 
-    public RootPlacer(IntProvider trunkOffsetY, Holder<BlockStateProvider> rootProvider, Optional<AboveRootPlacement> aboveRootPlacement) {
+    public RootPlacer(IntProvider trunkOffsetY, BlockStateProvider rootProvider, Optional<AboveRootPlacement> aboveRootPlacement) {
         this.trunkOffsetY = trunkOffsetY;
         this.rootProvider = rootProvider;
         this.aboveRootPlacement = aboveRootPlacement;
@@ -60,13 +_,13 @@
 
     protected void placeRoot(WorldGenLevel level, BiConsumer<BlockPos, BlockState> rootSetter, RandomSource random, BlockPos pos, TreeFeature tree) {
         if (this.canPlaceRoot(level, pos)) {
-            rootSetter.accept(pos, this.getPotentiallyWaterloggedState(level, pos, this.rootProvider.value().getState(level, random, pos)));
+            rootSetter.accept(pos, this.getPotentiallyWaterloggedState(level, pos, this.rootProvider.getState(level, random, pos)));
             if (this.aboveRootPlacement.isPresent()) {
                 AboveRootPlacement abovePlacement = this.aboveRootPlacement.get();
                 BlockPos above = pos.above();
                 if (random.nextFloat() < abovePlacement.aboveRootPlacementChance() && level.isStateAtPosition(above, BlockBehaviour.BlockStateBase::isAir)) {
                     rootSetter.accept(
-                        above, this.getPotentiallyWaterloggedState(level, above, abovePlacement.aboveRootProvider().value().getState(level, random, above))
+                        above, this.getPotentiallyWaterloggedState(level, above, abovePlacement.aboveRootProvider().getState(level, random, above))
                     );
                 }
             }
