--- a/net/minecraft/core/component/predicates/PotionsPredicate.java
+++ b/net/minecraft/core/component/predicates/PotionsPredicate.java
@@ -8,10 +_,10 @@
 import net.minecraft.advancements.predicates.SingleComponentItemPredicate;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.component.DataComponentType;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
@@ -23,7 +_,7 @@
     implements SingleComponentItemPredicate<PotionContents> {
     public static final Codec<PotionsPredicate> CODEC = RecordCodecBuilder.create(
         i -> i.group(
-                RegistryCodecs.holderSet(Registries.POTION).optionalFieldOf("potions").forGetter(PotionsPredicate::potions),
+                RegistryCodecs.homogeneousList(Registries.POTION).optionalFieldOf("potions").forGetter(PotionsPredicate::potions),
                 CollectionPredicate.<MobEffectInstance, MobEffectsPredicate>codec(MobEffectsPredicate.CODEC)
                     .optionalFieldOf("effects")
                     .forGetter(PotionsPredicate::effects)
