--- a/net/minecraft/core/component/predicates/TrimPredicate.java
+++ b/net/minecraft/core/component/predicates/TrimPredicate.java
@@ -5,10 +_,10 @@
 import java.util.Optional;
 import net.minecraft.advancements.predicates.SingleComponentItemPredicate;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.component.DataComponentType;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.world.item.equipment.trim.ArmorTrim;
 import net.minecraft.world.item.equipment.trim.TrimMaterial;
 import net.minecraft.world.item.equipment.trim.TrimPattern;
@@ -17,8 +_,8 @@
     implements SingleComponentItemPredicate<ArmorTrim> {
     public static final Codec<TrimPredicate> CODEC = RecordCodecBuilder.create(
         i -> i.group(
-                RegistryCodecs.holderSet(Registries.TRIM_MATERIAL).optionalFieldOf("material").forGetter(TrimPredicate::material),
-                RegistryCodecs.holderSet(Registries.TRIM_PATTERN).optionalFieldOf("pattern").forGetter(TrimPredicate::pattern)
+                RegistryCodecs.homogeneousList(Registries.TRIM_MATERIAL).optionalFieldOf("material").forGetter(TrimPredicate::material),
+                RegistryCodecs.homogeneousList(Registries.TRIM_PATTERN).optionalFieldOf("pattern").forGetter(TrimPredicate::pattern)
             )
             .apply(i, TrimPredicate::new)
     );
