--- a/net/minecraft/core/component/predicates/VillagerTypePredicate.java
+++ b/net/minecraft/core/component/predicates/VillagerTypePredicate.java
@@ -4,14 +_,14 @@
 import net.minecraft.advancements.predicates.SingleComponentItemPredicate;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.component.DataComponentType;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.world.entity.npc.villager.VillagerType;
 
 public record VillagerTypePredicate(HolderSet<VillagerType> villagerTypes) implements SingleComponentItemPredicate<Holder<VillagerType>> {
-    public static final Codec<VillagerTypePredicate> CODEC = RegistryCodecs.holderSet(Registries.VILLAGER_TYPE)
+    public static final Codec<VillagerTypePredicate> CODEC = RegistryCodecs.homogeneousList(Registries.VILLAGER_TYPE)
         .xmap(VillagerTypePredicate::new, VillagerTypePredicate::villagerTypes);
 
     @Override
