--- a/net/minecraft/world/level/border/WorldBorder.java
+++ b/net/minecraft/world/level/border/WorldBorder.java
@@ -6,9 +_,11 @@
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.BooleanOp;
@@ -28,6 +_,7 @@
     int absoluteMaxSize = 29999984;
     private WorldBorder.BorderExtent extent = new WorldBorder.StaticBorderExtent(5.999997E7F);
     public static final WorldBorder.Settings DEFAULT_SETTINGS = new WorldBorder.Settings(0.0, 0.0, 0.2, 5.0, 5, 15, 5.999997E7F, 0L, 0.0);
+    public ServerLevel world = null; // CraftBukkit
 
     public boolean isWithinBounds(BlockPos p_61938_) {
         return this.isWithinBounds((double)p_61938_.getX(), (double)p_61938_.getZ());
@@ -42,6 +_,18 @@
             && this.isWithinBounds((double)p_61928_.getMaxBlockX(), (double)p_61928_.getMaxBlockZ());
     }
 
+    // Paper start - Bound treasure maps to world border
+    private final BlockPos.MutableBlockPos mutPos = new BlockPos.MutableBlockPos();
+    public boolean isBlockInBounds(int chunkX, int chunkZ) {
+        this.mutPos.set(chunkX, 64, chunkZ);
+        return this.isWithinBounds(this.mutPos);
+    }
+    public boolean isChunkInBounds(int chunkX, int chunkZ) {
+        this.mutPos.set(((chunkX << 4) + 15), 64, (chunkZ << 4) + 15);
+        return this.isWithinBounds(this.mutPos);
+    }
+    // Paper end - Bound treasure maps to world border
+
     public boolean isWithinBounds(AABB p_61936_) {
         return this.isWithinBounds(p_61936_.minX, p_61936_.minZ, p_61936_.maxX - 1.0E-5F, p_61936_.maxZ - 1.0E-5F);
     }
@@ -127,6 +_,14 @@
     }
 
     public void setCenter(double p_61950_, double p_61951_) {
+        // Paper start - Add worldborder events
+        if (this.world != null) {
+            io.papermc.paper.event.world.border.WorldBorderCenterChangeEvent event = new io.papermc.paper.event.world.border.WorldBorderCenterChangeEvent(world.getWorld(), world.getWorld().getWorldBorder(), new org.bukkit.Location(world.getWorld(), this.getCenterX(), 0, this.getCenterZ()), new org.bukkit.Location(world.getWorld(), p_61950_, 0, p_61951_));
+            if (!event.callEvent()) return;
+            p_61950_ = event.getNewCenter().getX();
+            p_61951_ = event.getNewCenter().getZ();
+        }
+        // Paper end - Add worldborder events
         this.centerX = p_61950_;
         this.centerZ = p_61951_;
         this.extent.onCenterChange();
@@ -149,6 +_,17 @@
     }
 
     public void setSize(double p_61918_) {
+        // Paper start - Add worldborder events
+        if (this.world != null) {
+            io.papermc.paper.event.world.border.WorldBorderBoundsChangeEvent event = new io.papermc.paper.event.world.border.WorldBorderBoundsChangeEvent(world.getWorld(), world.getWorld().getWorldBorder(), io.papermc.paper.event.world.border.WorldBorderBoundsChangeEvent.Type.INSTANT_MOVE, getSize(), p_61918_, 0);
+            if (!event.callEvent()) return;
+            if (event.getType() == io.papermc.paper.event.world.border.WorldBorderBoundsChangeEvent.Type.STARTED_MOVE && event.getDuration() > 0) { // If changed to a timed transition
+                lerpSizeBetween(event.getOldSize(), event.getNewSize(), event.getDuration());
+                return;
+            }
+            p_61918_ = event.getNewSize();
+        }
+        // Paper end - Add worldborder events
         this.extent = new WorldBorder.StaticBorderExtent(p_61918_);
 
         for (BorderChangeListener borderchangelistener : this.getListeners()) {
@@ -157,6 +_,20 @@
     }
 
     public void lerpSizeBetween(double p_61920_, double p_61921_, long p_61922_) {
+        // Paper start - Add worldborder events
+        if (this.world != null) {
+            io.papermc.paper.event.world.border.WorldBorderBoundsChangeEvent.Type type;
+            if (p_61920_ == p_61921_) { // new size = old size
+                type = io.papermc.paper.event.world.border.WorldBorderBoundsChangeEvent.Type.INSTANT_MOVE; // Use INSTANT_MOVE because below it creates a Static border if they are equal.
+            } else {
+                type = io.papermc.paper.event.world.border.WorldBorderBoundsChangeEvent.Type.STARTED_MOVE;
+            }
+            io.papermc.paper.event.world.border.WorldBorderBoundsChangeEvent event = new io.papermc.paper.event.world.border.WorldBorderBoundsChangeEvent(world.getWorld(), world.getWorld().getWorldBorder(), type, p_61920_, p_61921_, p_61922_);
+            if (!event.callEvent()) return;
+            p_61921_ = event.getNewSize();
+            p_61922_ = event.getDuration();
+        }
+        // Paper end - Add worldborder events
         this.extent = (WorldBorder.BorderExtent)(p_61920_ == p_61921_
             ? new WorldBorder.StaticBorderExtent(p_61921_)
             : new WorldBorder.MovingBorderExtent(p_61920_, p_61921_, p_61922_));
@@ -171,6 +_,7 @@
     }
 
     public void addListener(BorderChangeListener p_61930_) {
+        if (listeners.contains(p_61930_)) return; // CraftBukkit
         this.listeners.add(p_61930_);
     }
 
@@ -367,6 +_,7 @@
 
         @Override
         public WorldBorder.BorderExtent update() {
+            if (world != null && this.getLerpRemainingTime() <= 0L) new io.papermc.paper.event.world.border.WorldBorderBoundsChangeFinishEvent(world.getWorld(), world.getWorld().getWorldBorder(), this.from, this.to, this.lerpDuration).callEvent(); // Paper - Add worldborder events
             return (WorldBorder.BorderExtent)(this.getLerpRemainingTime() <= 0L ? WorldBorder.this.new StaticBorderExtent(this.to) : this);
         }
 
