--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -25,6 +_,7 @@
 import net.minecraft.world.flag.FeatureFlags;
 import net.minecraft.world.item.CreativeModeTabs;
 import net.minecraft.world.item.Item;
+import net.minecraft.world.level.block.ComposterBlock;
 import net.minecraft.world.level.block.FireBlock;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.gamerules.GameRule;
@@ -43,11 +_,13 @@
         if (!isBootstrapped) {
             isBootstrapped = true;
             Instant start = Instant.now();
+            io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enterBootstrappers(); // Paper - Entrypoint for bootstrapping
             if (BuiltInRegistries.REGISTRY.keySet().isEmpty()) {
                 throw new IllegalStateException("Unable to load registries");
             }
 
             FireBlock.bootStrap();
+            ComposterBlock.bootStrap();
             if (EntityType.getKey(EntityTypes.PLAYER) == null) {
                 throw new IllegalStateException("Failed loading EntityTypes");
             }
@@ -55,8 +_,14 @@
             EntitySelectorOptions.bootStrap();
             DispenseItemBehavior.bootStrap();
             CauldronInteractions.bootStrap();
-            BuiltInRegistries.bootStrap();
+            // Paper start
+            BuiltInRegistries.bootStrap(() -> {
+                io.papermc.paper.world.worldgen.OptionallyFlatBedrockConditionSource.bootstrap(); // Paper - Flat bedrock generator settings
+            });
+            // Paper end
             CreativeModeTabs.validate();
+            net.neoforged.neoforge.registries.GameData.vanillaSnapshot();
+            if (false) // skip redirectOutputToLog, Forge already redirects stdout and stderr output to log so that they print with more context
             wrapStreams();
             bootstrapDuration.set(Duration.between(start, Instant.now()).toMillis());
         }
@@ -119,7 +_,6 @@
             Commands.validate();
         }
 
-        DefaultAttributes.validate();
     }
 
     private static void wrapStreams() {
