--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -44,6 +_,7 @@
         if (!isBootstrapped) {
             isBootstrapped = true;
             Instant start = Instant.now();
+            io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enterBootstrappers(); // Paper - Entrypoint for bootstrapping
             if (BuiltInRegistries.REGISTRY.keySet().isEmpty()) {
                 throw new IllegalStateException("Unable to load registries");
             }
@@ -56,9 +_,15 @@
             EntitySelectorOptions.bootStrap();
             DispenseItemBehavior.bootStrap();
             CauldronInteractions.bootStrap();
-            BuiltInRegistries.bootStrap();
+            // Paper start
+            BuiltInRegistries.bootStrap(() -> {
+                io.papermc.paper.world.worldgen.OptionallyFlatBedrockConditionSource.bootstrap(); // Paper - Flat bedrock generator settings
+            });
+            // Paper end
             CreativeModeTabs.validate();
             LootContextParamSets.validate();
+            net.neoforged.neoforge.registries.GameData.vanillaSnapshot();
+            if (false) // skip redirectOutputToLog, Forge already redirects stdout and stderr output to log so that they print with more context
             wrapStreams();
             bootstrapDuration.set(Duration.between(start, Instant.now()).toMillis());
         }
@@ -121,7 +_,6 @@
             Commands.validate();
         }
 
-        DefaultAttributes.validate();
     }
 
     private static void wrapStreams() {
