--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -39,6 +_,7 @@
         if (!isBootstrapped) {
             isBootstrapped = true;
             Instant instant = Instant.now();
+            io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enterBootstrappers(); // Paper - Entrypoint for bootstrapping
             if (BuiltInRegistries.REGISTRY.keySet().isEmpty()) {
                 throw new IllegalStateException("Unable to load registries");
             } else {
@@ -52,6 +_,8 @@
                     CauldronInteraction.bootStrap();
                     BuiltInRegistries.bootStrap();
                     CreativeModeTabs.validate();
+                    net.neoforged.neoforge.registries.GameData.vanillaSnapshot();
+                    if (false) // skip redirectOutputToLog, Forge already redirects stdout and stderr output to log so that they print with more context
                     wrapStreams();
                     bootstrapDuration.set(Duration.between(instant, Instant.now()).toMillis());
                 }
@@ -117,7 +_,6 @@
             Commands.validate();
         }
 
-        DefaultAttributes.validate();
     }
 
     private static void wrapStreams() {
