--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -1,6 +_,6 @@
 package net.minecraft.server;
 
-import com.mojang.authlib.services.MinecraftServicesDiscoveryService;
+import com.mohistmc.youer.Youer;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Dynamic;
@@ -13,7 +_,10 @@
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BooleanSupplier;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
@@ -36,6 +_,7 @@
 import net.minecraft.server.jsonrpc.JsonRpc;
 import net.minecraft.server.jsonrpc.ManagementServer;
 import net.minecraft.server.notifications.NotificationManager;
+import net.minecraft.server.packs.PackType;
 import net.minecraft.server.packs.repository.PackRepository;
 import net.minecraft.server.packs.repository.ServerPacksSource;
 import net.minecraft.util.Mth;
@@ -56,21 +_,26 @@
 import net.minecraft.world.level.levelgen.presets.WorldPresets;
 import net.minecraft.world.level.storage.LevelDataAndDimensions;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import net.minecraft.world.level.storage.LevelResource;
 import net.minecraft.world.level.storage.LevelSummary;
 import net.minecraft.world.level.storage.PrimaryLevelData;
-import net.minecraft.world.level.storage.WorldData;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+import static java.util.Arrays.asList;
+
 public class Main {
     private static final Logger LOGGER = LogUtils.getLogger();
 
     @SuppressForbidden(reason = "System.out needed before bootstrap")
     public static void main(String[] args) {
+        Youer.initProperty();
+        io.papermc.paper.log.LogManagerShutdownThread.hook(); // Paper - Improved watchdog support
+        Youer.initI18n(); // Youer
         SharedConstants.tryDetectVersion();
         OptionParser parser = new OptionParser();
         OptionSpec<Void> nogui = parser.accepts("nogui");
-        OptionSpec<Void> initSettings = parser.accepts("initSettings", "Initializes 'server.properties' and 'eula.txt', then quits");
+        OptionSpec<Void> initSettings = parser.accepts("initSettings", "Only create configuration files and then exit");// SPIGOT-5761: Add initSettings option
         OptionSpec<Void> demo = parser.accepts("demo");
         OptionSpec<Void> bonusChest = parser.accepts("bonusChest");
         OptionSpec<Void> forceUpgrade = parser.accepts("forceUpgrade");
@@ -86,12 +_,109 @@
         OptionSpec<Path> pidFile = parser.accepts("pidFile").withRequiredArg().withValuesConvertedBy(new PathConverter());
         OptionSpec<String> nonOptions = parser.nonOptions();
 
+        parser.acceptsAll(Arrays.asList("c", "config"), "Properties file to use")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("server.properties"))
+                .describedAs("Properties file");
+        parser.acceptsAll(Arrays.asList("b", "bukkit-settings"), "File for bukkit settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("bukkit.yml"))
+                .describedAs("Yml file");
+
+        parser.acceptsAll(Arrays.asList("C", "commands-settings"), "File for command settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("commands.yml"))
+                .describedAs("Yml file");
+
+        parser.acceptsAll(Arrays.asList("P", "plugins"), "Plugin directory to use")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("plugins"))
+                .describedAs("Plugin directory");
+
+        parser.acceptsAll(List.of("nojline"), "Disables jline and emulates the vanilla console");
+        parser.acceptsAll(List.of("noconsole"), "Disables the console");
+
+        // Spigot Start
+        parser.acceptsAll(Arrays.asList("S", "spigot-settings"), "File for spigot settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("spigot.yml"))
+                .describedAs("Yml file");
+        // Spigot End
+
+        parser.acceptsAll(asList("paper-dir", "paper-settings-directory"), "Directory for Paper settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File(io.papermc.paper.configuration.PaperConfigurations.CONFIG_DIR))
+                .describedAs("Config directory");
+
+        parser.acceptsAll(asList("paper", "paper-settings"), "File for Paper settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("paper.yml"))
+                .describedAs("Yml file");
+
+        parser.acceptsAll(asList("add-plugin", "add-extra-plugin-jar"), "Specify paths to extra plugin jars to be loaded in addition to those in the plugins folder. This argument can be specified multiple times, once for each extra plugin jar path.")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File[] {})
+                .describedAs("Jar file");
+
+        parser.acceptsAll(asList("add-plugin-dir", "add-extra-plugin-dir"), "Specify paths to extra plugin directories to be loaded in addition to the plugins folder. This argument can be specified multiple times, once for each extra plugin dir path.")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File[] {})
+                .describedAs("Plugin directory");
+
+        parser.accepts("server-name", "Name of the server")
+                .withRequiredArg()
+                .ofType(String.class)
+                .defaultsTo("Unknown Server")
+                .describedAs("Name");
+
+        // Purpur start - Purpur config files
+        parser.acceptsAll(Arrays.asList("purpur", "purpur-settings"), "File for purpur settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("purpur.yml"))
+                .describedAs("Yml file");
+        // Purpur end - Purpur config files
+
+        // Youer Start
+        parser.acceptsAll(Arrays.asList("Y", "youer-settings"), "File for youer settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("youer-config","youer.yml"))
+                .describedAs("Yml file");
+        // Youer End
+
+        OptionSet options = parser.parse(args);
         try {
-            OptionSet options = parser.parse(args);
             if (options.has(help)) {
                 parser.printHelpOn(System.err);
                 return;
             }
+            if (options.has("nojline")) {
+                System.setProperty(net.minecrell.terminalconsole.TerminalConsoleAppender.JLINE_OVERRIDE_PROPERTY, "false");
+                org.bukkit.craftbukkit.Main.useJline = false;
+            }
+
+            if (options.has("noconsole")) {
+                org.bukkit.craftbukkit.Main.useConsole = false;
+                org.bukkit.craftbukkit.Main.useJline = false; // Paper
+                System.setProperty(net.minecrell.terminalconsole.TerminalConsoleAppender.JLINE_OVERRIDE_PROPERTY, "false"); // Paper
+            }
+            Path eulaFile = Paths.get("eula.txt");
+            Eula eula = new Eula(eulaFile);
+
+            // Paper start - load config files early for access below if needed
+            org.bukkit.configuration.file.YamlConfiguration bukkitConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) options.valueOf("bukkit-settings"));
+            org.bukkit.configuration.file.YamlConfiguration spigotConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) options.valueOf("spigot-settings"));
+            // Paper end - load config files early for access below if needed
 
             Path pidFilePath = options.valueOf(pidFile);
             if (pidFilePath != null) {
@@ -103,30 +_,98 @@
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            // Purpur start - Add toggle for enchant level clamping - load config files early
+            org.bukkit.configuration.file.YamlConfiguration purpurConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) options.valueOf("purpur-settings"));
+            org.purpurmc.purpur.PurpurConfig.clampEnchantLevels = purpurConfiguration.getBoolean("settings.enchantment.clamp-levels", true);
+            org.purpurmc.purpur.PurpurConfig.registerMinecraftDebugCommands = purpurConfiguration.getBoolean("settings.register-minecraft-debug-commands"); // Purpur - register minecraft debug commands
+            org.purpurmc.purpur.PurpurConfig.registerMinecraftDisabledCommands = purpurConfiguration.getBoolean("settings.register-minecraft-disabled-commands"); // Purpur - register disabled minecraft commands
+            // Purpur end - Add toggle for enchant level clamping - load config files early
+
+            io.papermc.paper.plugin.PluginInitializerManager.load(options); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
             Util.startTimerHackThread();
             Path settingsFile = Paths.get("server.properties");
-            DedicatedServerSettings settings = new DedicatedServerSettings(settingsFile);
+            if (!options.has(initSettings)) net.neoforged.neoforge.server.loading.ServerModLoader.load(false); // Load mods before we load almost anything else anymore. Single spot now. Only loads if they haven't passed the initserver param
+            DedicatedServerSettings settings = new DedicatedServerSettings(options);// CraftBukkit - CLI argument support
             settings.forceSave();
+            // Paper start
+            if (options.has("forceUpgrade") || options.has("recreateRegionFiles")) {
+                LOGGER.error("World upgrade and region file recreation are not yet implemented in Paper 26.1.");
+                return;
+            }
+            // Paper end
             RegionFileVersion.configure(settings.getProperties().regionFileComression);
-            Path eulaFile = Paths.get("eula.txt");
-            Eula eula = new Eula(eulaFile);
             if (options.has(initSettings)) {
+                // CraftBukkit start - SPIGOT-5761: Create bukkit.yml and commands.yml if not present
+                File configFile = (File) options.valueOf("bukkit-settings");
+                org.bukkit.configuration.file.YamlConfiguration configuration = org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(configFile);
+                configuration.options().copyDefaults(true);
+                configuration.setDefaults(org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(new java.io.InputStreamReader(Main.class.getClassLoader().getResourceAsStream("configurations/bukkit.yml"), java.nio.charset.StandardCharsets.UTF_8)));
+                configuration.save(configFile);
+
+                File commandFile = (File) options.valueOf("commands-settings");
+                org.bukkit.configuration.file.YamlConfiguration commandsConfiguration = org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(commandFile);
+                commandsConfiguration.options().copyDefaults(true);
+                commandsConfiguration.setDefaults(org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(new java.io.InputStreamReader(Main.class.getClassLoader().getResourceAsStream("configurations/commands.yml"), java.nio.charset.StandardCharsets.UTF_8)));
+                commandsConfiguration.save(commandFile);
+                // CraftBukkit end
                 LOGGER.info("Initialized '{}' and '{}'", settingsFile.toAbsolutePath(), eulaFile.toAbsolutePath());
                 return;
             }
 
-            if (!eula.hasAgreedToEULA()) {
+            // Paper start - eula system property
+            boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
+            if (eulaAgreed) {
+                LOGGER.warn("You have used the Paper command line EULA agreement flag.");
+                LOGGER.warn("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
+                LOGGER.warn("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
+            } else if (!eula.hasAgreedToEULA()) {
+                // Paper end - eula system property
                 LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
                 return;
             }
 
-            File universePath = new File(options.valueOf(universe));
-            Services services = Services.create(MinecraftServicesDiscoveryService.create(Proxy.NO_PROXY), universePath);
+            // Paper start - Detect headless JRE
+            String awtException = io.papermc.paper.util.ServerEnvironment.awtDependencyCheck();
+            if (awtException != null) {
+                LOGGER.error("You are using a headless JRE distribution.");
+                LOGGER.error("This distribution is missing certain graphic libraries that the Minecraft server needs to function.");
+                LOGGER.error("For instructions on how to install the non-headless JRE, see https://docs.papermc.io/misc/java-install");
+                LOGGER.error("");
+                LOGGER.error(awtException);
+                return;
+            }
+            // Paper end - Detect headless JRE
+
+            File configYmenu = new File("youer-config/menu", "example.yml");
+            if (!configYmenu.exists()) {
+                org.bukkit.configuration.file.YamlConfiguration configurationYmenu = org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(configYmenu);
+                configurationYmenu.options().copyDefaults(true);
+                configurationYmenu.setDefaults(org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(new java.io.InputStreamReader(Main.class.getClassLoader().getResourceAsStream("youer-config/menu/example.yml"), com.google.common.base.Charsets.UTF_8)));
+                configurationYmenu.save(configYmenu);
+            }
+
+            org.spigotmc.SpigotConfig.disabledAdvancements = spigotConfiguration.getStringList("advancements.disabled"); // Paper - fix SPIGOT-5885, must be set early in init
+
+            // Paper start - fix SPIGOT-5824
+            File universePath;
+            File userCacheFile = new File(Services.USERID_CACHE_FILE);
+            if (options.has("universe")) {
+                universePath = (File) options.valueOf("universe"); // CraftBukkit
+                userCacheFile = new File(universePath, Services.USERID_CACHE_FILE);
+            } else {
+                universePath = new File(bukkitConfiguration.getString("settings.world-container", "."));
+            }
+            // Paper end - fix SPIGOT-5824
+            Services services = Services.create(new com.destroystokyo.paper.profile.PaperAuthenticationService(Proxy.NO_PROXY), universePath, userCacheFile, options); // Paper - pass OptionSet to load paper config files; override authentication service; fix world-container
             NotificationManager notificationManager = new NotificationManager();
             ManagementServer jsonRpcServer = JsonRpc.create(settings, notificationManager);
-            String levelName = Optional.ofNullable(options.valueOf(worldName)).orElse(settings.getProperties().levelName);
+            String levelName = Optional.ofNullable((String) options.valueOf("world")).orElse(settings.getProperties().levelName);
+            if (levelName == null || levelName.isEmpty() || new File(universePath, levelName).getAbsolutePath().equals(new File(levelName).getAbsolutePath())) {
+                LOGGER.error("Invalid world directory specified, must not be null, empty or the same directory as your universe! {}", levelName);
+                return;
+            }
             LevelStorageSource levelStorageSource = LevelStorageSource.createDefault(universePath.toPath());
             LevelStorageSource.LevelStorageAccess access = levelStorageSource.validateAndCreateAccess(levelName);
             Dynamic<?> levelDataTag;
@@ -150,10 +_,11 @@
                     return;
                 }
 
-                levelDataTag = DataFixers.getFileFixer().fix(access, levelDataUnfixed, new UpgradeProgress(notificationManager));
+                levelDataTag = DataFixers.getFileFixer().fix(access, levelDataUnfixed, new UpgradeProgress());
             } else {
                 levelDataTag = null;
             }
+            io.papermc.paper.world.migration.WorldFolderMigration.didInitialLoad = true; // Paper
 
             boolean safeModeEnabled = options.has(safeMode);
             if (safeModeEnabled) {
@@ -161,6 +_,25 @@
             }
 
             PackRepository packRepository = ServerPacksSource.createPackRepository(access);
+            // CraftBukkit start
+            File bukkitDataPackFolder = new File(access.getLevelPath(LevelResource.DATAPACK_DIR).toFile(), "bukkit");
+            if (!bukkitDataPackFolder.exists()) {
+                bukkitDataPackFolder.mkdirs();
+            }
+            File mcMeta = new File(bukkitDataPackFolder, "pack.mcmeta");
+            try {
+                com.google.common.io.Files.write("{\n"
+                        + "    \"pack\": {\n"
+                        + "        \"description\": \"Data pack for resources provided by Bukkit plugins\",\n"
+                        + "        \"min_format\": " + SharedConstants.getCurrentVersion().packVersion(PackType.SERVER_DATA) + ",\n"
+                        + "        \"max_format\": " + SharedConstants.getCurrentVersion().packVersion(PackType.SERVER_DATA) + "\n"
+                        + "    }\n"
+                        + "}\n", mcMeta, com.google.common.base.Charsets.UTF_8);
+            } catch (java.io.IOException ex) {
+                throw new RuntimeException("Could not initialize Bukkit datapack", ex);
+            }
+            AtomicReference<WorldLoader.DataLoadContext> worldLoader = new AtomicReference<>();
+            // CraftBukkit end
 
             WorldStem worldStem;
             try {
@@ -169,13 +_,27 @@
                         executor -> WorldLoader.load(
                             worldLoadConfig,
                             context -> {
+                                worldLoader.set(context); // CraftBukkit
                                 Registry<LevelStem> datapackDimensions = context.datapackDimensions().lookupOrThrow(Registries.LEVEL_STEM);
                                 if (levelDataTag != null) {
+                                    // Paper start - migrate startup world
+                                    try {
+                                        io.papermc.paper.world.migration.WorldFolderMigration.migrateStartupWorld(
+                                                access,
+                                                context.datapackWorldgen(),
+                                                levelName,
+                                                LevelStem.OVERWORLD,
+                                                Registries.levelStemToLevel(LevelStem.OVERWORLD)
+                                        );
+                                    } catch (final IOException ex) {
+                                        throw new UncheckedIOException("Failed to migrate world storage for " + LevelStem.OVERWORLD.identifier(), ex);
+                                    }
+                                    // Paper end - migrate startup world
                                     LevelDataAndDimensions worldData = LevelStorageSource.getLevelDataAndDimensions(
-                                        access, levelDataTag, context.dataConfiguration(), datapackDimensions, context.datapackWorldgen()
+                                            access, levelDataTag, context.dataConfiguration(), datapackDimensions, context.datapackWorldgen(), net.minecraft.world.level.Level.OVERWORLD // Paper
                                     );
                                     return new WorldLoader.DataLoadOutput<>(
-                                        worldData.worldDataAndGenSettings(), worldData.dimensions().dimensionsRegistryAccess()
+                                            worldData.worldDataAndGenSettings(), worldData.dimensions().dimensionsRegistryAccess()
                                     );
                                 } else {
                                     LOGGER.info("No existing world data, creating new world");
@@ -194,7 +_,7 @@
                 );
                 return;
             }
-
+            /*
             RegistryAccess.Frozen registryHolder = worldStem.registries().compositeAccess();
             WorldData data = worldStem.worldDataAndGenSettings().data();
             boolean recreateRegionFilesValue = options.has(recreateRegionFiles);
@@ -203,9 +_,12 @@
             }
 
             access.saveDataTag(data);
+            */
             final DedicatedServer dedicatedServer = MinecraftServer.spin(
                 thread -> {
                     DedicatedServer server = new DedicatedServer(
+                        options,
+                        worldLoader.get(),
                         thread,
                         access,
                         packRepository,
@@ -229,20 +_,23 @@
                     return server;
                 }
             );
+            /* CraftBukkit start
             Thread shutdownThread = new Thread("Server Shutdown Thread") {
                 @Override
                 public void run() {
                     dedicatedServer.halt(true);
+                    org.apache.logging.log4j.LogManager.shutdown(); // we're manually managing the logging shutdown on the server. Make sure we do it here at the end.
                 }
             };
             shutdownThread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
             Runtime.getRuntime().addShutdownHook(shutdownThread);
+            */ // CraftBukkit end
         } catch (Throwable t) {
             LOGGER.error(LogUtils.FATAL_MARKER, "Failed to start the minecraft server", t);
         }
     }
 
-    private static WorldLoader.DataLoadOutput<LevelDataAndDimensions.WorldDataAndGenSettings> createNewWorldData(
+    public static WorldLoader.DataLoadOutput<LevelDataAndDimensions.WorldDataAndGenSettings> createNewWorldData(
         DedicatedServerSettings settings, WorldLoader.DataLoadContext context, Registry<LevelStem> datapackDimensions, boolean demoMode, boolean bonusChest
     ) {
         LevelSettings createLevelSettings;
@@ -265,6 +_,9 @@
             dimensions = properties.createDimensions(context.datapackWorldgen());
         }
 
+        // Neo: Do a write-read-cycle to inject modded dimensions on first start of a dedicated server into its generated world dimensions list.
+        var registryOps = net.minecraft.resources.RegistryOps.create(net.minecraft.nbt.NbtOps.INSTANCE, context.datapackWorldgen());
+        dimensions = WorldDimensions.CODEC.encoder().encodeStart(registryOps, dimensions).flatMap((writtenPayloadWithModdedDimensions) -> WorldDimensions.CODEC.decoder().parse(registryOps, writtenPayloadWithModdedDimensions)).resultOrPartial(LOGGER::error).orElse(dimensions);
         WorldDimensions.Complete finalDimensions = dimensions.bake(datapackDimensions);
         Lifecycle lifecycle = finalDimensions.lifecycle().add(context.datapackWorldgen().allRegistriesLifecycle());
         PrimaryLevelData primaryLevelData = new PrimaryLevelData(createLevelSettings, finalDimensions.specialWorldProperty(), lifecycle);
@@ -309,7 +_,8 @@
         RegistryAccess registryAccess,
         boolean recreateRegionFiles
     ) {
-        LOGGER.info("Forcing world upgrade!");
+        /*
+        LOGGER.info("Forcing world upgrade! {}", storageSource.getLevelId()); // CraftBukkit
 
         try (WorldUpgrader upgrader = new WorldUpgrader(storageSource, fixerUpper, registryAccess, eraseCache, recreateRegionFiles)) {
             Component lastStatus = null;
@@ -336,6 +_,6 @@
                     }
                 }
             }
-        }
+        }*/
     }
 }
