--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -1,5 +_,7 @@
 package net.minecraft.server;
 
+import com.google.common.base.Charsets;
+import com.mohistmc.youer.Youer;
 import com.mojang.authlib.yggdrasil.YggdrasilAuthenticationService;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
@@ -8,12 +_,16 @@
 import java.awt.GraphicsEnvironment;
 import java.io.File;
 import java.io.IOException;
+import java.io.InputStreamReader;
 import java.io.UncheckedIOException;
 import java.net.Proxy;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BooleanSupplier;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
@@ -33,6 +_,7 @@
 import net.minecraft.server.dedicated.DedicatedServer;
 import net.minecraft.server.dedicated.DedicatedServerProperties;
 import net.minecraft.server.dedicated.DedicatedServerSettings;
+import net.minecraft.server.packs.PackType;
 import net.minecraft.server.packs.repository.PackRepository;
 import net.minecraft.server.packs.repository.ServerPacksSource;
 import net.minecraft.util.Mth;
@@ -53,9 +_,11 @@
 import net.minecraft.world.level.levelgen.presets.WorldPresets;
 import net.minecraft.world.level.storage.LevelDataAndDimensions;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import net.minecraft.world.level.storage.LevelResource;
 import net.minecraft.world.level.storage.LevelSummary;
 import net.minecraft.world.level.storage.PrimaryLevelData;
 import net.minecraft.world.level.storage.WorldData;
+import org.bukkit.configuration.file.YamlConfiguration;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -64,10 +_,11 @@
 
     @SuppressForbidden(reason = "System.out needed before bootstrap")
     public static void main(String[] args) {
+        Youer.initI18n(); // Youer
         SharedConstants.tryDetectVersion();
         OptionParser parser = new OptionParser();
         OptionSpec<Void> nogui = parser.accepts("nogui");
-        OptionSpec<Void> initSettings = parser.accepts("initSettings", "Initializes 'server.properties' and 'eula.txt', then quits");
+        OptionSpec<Void> initSettings = parser.accepts("initSettings", "Only create configuration files and then exit");// SPIGOT-5761: Add initSettings option
         OptionSpec<Void> demo = parser.accepts("demo");
         OptionSpec<Void> bonusChest = parser.accepts("bonusChest");
         OptionSpec<Void> forceUpgrade = parser.accepts("forceUpgrade");
@@ -83,12 +_,70 @@
         OptionSpec<Path> pidFile = parser.accepts("pidFile").withRequiredArg().withValuesConvertedBy(new PathConverter());
         OptionSpec<String> nonOptions = parser.nonOptions();
 
+        parser.acceptsAll(Arrays.asList("c", "config"), "Properties file to use")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("server.properties"))
+                .describedAs("Properties file");
+        parser.acceptsAll(Arrays.asList("b", "bukkit-settings"), "File for bukkit settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("bukkit.yml"))
+                .describedAs("Yml file");
+
+        parser.acceptsAll(Arrays.asList("C", "commands-settings"), "File for command settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("commands.yml"))
+                .describedAs("Yml file");
+
+        parser.acceptsAll(Arrays.asList("P", "plugins"), "Plugin directory to use")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("plugins"))
+                .describedAs("Plugin directory");
+
+        parser.acceptsAll(List.of("nojline"), "Disables jline and emulates the vanilla console");
+        parser.acceptsAll(List.of("noconsole"), "Disables the console");
+
+        // Spigot Start
+        parser.acceptsAll(Arrays.asList("S", "spigot-settings"), "File for spigot settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("spigot.yml"))
+                .describedAs("Yml file");
+        // Spigot End
+
+        // Youer Start
+        parser.acceptsAll(Arrays.asList("Y", "youer-settings"), "File for youer settings")
+                .withRequiredArg()
+                .ofType(File.class)
+                .defaultsTo(new File("youer-config","youer.yml"))
+                .describedAs("Yml file");
+        // Youer End
+
+        OptionSet options = parser.parse(args);
         try {
-            OptionSet options = parser.parse(args);
             if (options.has(help)) {
                 parser.printHelpOn(System.err);
                 return;
             }
+            Path eulaFile = Paths.get("eula.txt");
+            Eula eula = new Eula(eulaFile);
+
+            // Spigot Start
+            boolean eulaAgreed = Boolean.getBoolean( "com.mojang.eula.agree" );
+            if ( eulaAgreed )
+            {
+                System.err.println( "You have used the Spigot command line EULA agreement flag." );
+                System.err.println( "By using this setting you are indicating your agreement to Mojang's EULA (https://account.mojang.com/documents/minecraft_eula)." );
+                System.err.println( "If you do not agree to the above EULA please stop your server and remove this flag immediately." );
+            }
+            // Spigot End
+            if (!eula.hasAgreedToEULA() && !eulaAgreed) { // Spigot
+                LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+                return;
+            }
 
             Path pidFilePath = options.valueOf(pidFile);
             if (pidFilePath != null) {
@@ -104,26 +_,45 @@
             Bootstrap.validate();
             Util.startTimerHackThread();
             Path settingsFile = Paths.get("server.properties");
-            DedicatedServerSettings settings = new DedicatedServerSettings(settingsFile);
+            if (!options.has(initSettings)) net.neoforged.neoforge.server.loading.ServerModLoader.load(false); // Load mods before we load almost anything else anymore. Single spot now. Only loads if they haven't passed the initserver param
+            DedicatedServerSettings settings = new DedicatedServerSettings(options);// CraftBukkit - CLI argument support
             settings.forceSave();
             RegionFileVersion.configure(settings.getProperties().regionFileComression);
-            Path eulaFile = Paths.get("eula.txt");
-            Eula eula = new Eula(eulaFile);
             if (options.has(initSettings)) {
+                // CraftBukkit start - SPIGOT-5761: Create bukkit.yml and commands.yml if not present
+                File configFile = (File) options.valueOf("bukkit-settings");
+                YamlConfiguration configuration = YamlConfiguration.loadConfiguration(configFile);
+                configuration.options().copyDefaults(true);
+                configuration.setDefaults(YamlConfiguration.loadConfiguration(new InputStreamReader(Main.class.getClassLoader().getResourceAsStream("configurations/bukkit.yml"), Charsets.UTF_8)));
+                configuration.save(configFile);
+
+                File commandFile = (File) options.valueOf("commands-settings");
+                YamlConfiguration commandsConfiguration = YamlConfiguration.loadConfiguration(commandFile);
+                commandsConfiguration.options().copyDefaults(true);
+                commandsConfiguration.setDefaults(YamlConfiguration.loadConfiguration(new InputStreamReader(Main.class.getClassLoader().getResourceAsStream("configurations/commands.yml"), Charsets.UTF_8)));
+                commandsConfiguration.save(commandFile);
+                // CraftBukkit end
                 LOGGER.info("Initialized '{}' and '{}'", settingsFile.toAbsolutePath(), eulaFile.toAbsolutePath());
                 return;
             }
 
-            if (!eula.hasAgreedToEULA()) {
-                LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
-                return;
+            File configYmenu = new File("youer-config/menu", "example.yml");
+            if (!configYmenu.exists()) {
+                org.bukkit.configuration.file.YamlConfiguration configurationYmenu = org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(configYmenu);
+                configurationYmenu.options().copyDefaults(true);
+                configurationYmenu.setDefaults(org.bukkit.configuration.file.YamlConfiguration.loadConfiguration(new java.io.InputStreamReader(Main.class.getClassLoader().getResourceAsStream("youer-config/menu/example.yml"), com.google.common.base.Charsets.UTF_8)));
+                configurationYmenu.save(configYmenu);
             }
 
             File universePath = new File(options.valueOf(universe));
             Services services = Services.create(new YggdrasilAuthenticationService(Proxy.NO_PROXY), universePath);
             String levelName = Optional.ofNullable(options.valueOf(worldName)).orElse(settings.getProperties().levelName);
+            if (levelName == null || levelName.isEmpty() || new File(universePath, levelName).getAbsolutePath().equals(new File(levelName).getAbsolutePath())) {
+                LOGGER.error("Invalid world directory specified, must not be null, empty or the same directory as your universe! {}", levelName);
+                return;
+            }
             LevelStorageSource levelStorageSource = LevelStorageSource.createDefault(universePath.toPath());
-            LevelStorageSource.LevelStorageAccess access = levelStorageSource.validateAndCreateAccess(levelName);
+            LevelStorageSource.LevelStorageAccess access = levelStorageSource.validateAndCreateAccess(levelName, LevelStem.OVERWORLD);
             Dynamic<?> levelDataTag;
             if (access.hasWorldData()) {
                 Dynamic<?> levelDataUnfixed;
@@ -156,6 +_,26 @@
             }
 
             PackRepository packRepository = ServerPacksSource.createPackRepository(access);
+            // CraftBukkit start
+            File bukkitDataPackFolder = new File(access.getLevelPath(LevelResource.DATAPACK_DIR).toFile(), "bukkit");
+            if (!bukkitDataPackFolder.exists()) {
+                bukkitDataPackFolder.mkdirs();
+            }
+            File mcMeta = new File(bukkitDataPackFolder, "pack.mcmeta");
+            try {
+                com.google.common.io.Files.write("{\n"
+                        + "    \"pack\": {\n"
+                        + "        \"description\": \"Data pack for resources provided by Bukkit plugins\",\n"
+                        + "        \"min_format\": " + SharedConstants.getCurrentVersion().packVersion(PackType.SERVER_DATA) + ",\n"
+                        + "        \"max_format\": " + SharedConstants.getCurrentVersion().packVersion(PackType.SERVER_DATA) + "\n"
+                        + "    }\n"
+                        + "}\n", mcMeta, com.google.common.base.Charsets.UTF_8);
+            } catch (java.io.IOException ex) {
+                throw new RuntimeException("Could not initialize Bukkit datapack", ex);
+            }
+            // CraftBukkit end
+
+            AtomicReference<WorldLoader.DataLoadContext> worldLoader = new AtomicReference<>();
 
             WorldStem worldStem;
             try {
@@ -164,6 +_,7 @@
                         executor -> WorldLoader.load(
                             worldLoadConfig,
                             context -> {
+                                worldLoader.set(context); // CraftBukkit
                                 Registry<LevelStem> datapackDimensions = context.datapackDimensions().lookupOrThrow(Registries.LEVEL_STEM);
                                 if (levelDataTag != null) {
                                     LevelDataAndDimensions worldData = LevelStorageSource.getLevelDataAndDimensions(
@@ -190,7 +_,7 @@
                 );
                 return;
             }
-
+            /*
             RegistryAccess.Frozen registryHolder = worldStem.registries().compositeAccess();
             WorldData data = worldStem.worldDataAndGenSettings().data();
             boolean recreateRegionFilesValue = options.has(recreateRegionFiles);
@@ -199,9 +_,11 @@
             }
 
             access.saveDataTag(data);
+            */
             final DedicatedServer dedicatedServer = MinecraftServer.spin(
                 thread -> {
                     DedicatedServer server = new DedicatedServer(
+                        options, worldLoader.get(),
                         thread, access, packRepository, worldStem, Optional.empty(), settings, DataFixers.getDataFixer(), services
                     );
                     server.setPort(options.valueOf(port));
@@ -215,14 +_,17 @@
                     return server;
                 }
             );
+            /* CraftBukkit start
             Thread shutdownThread = new Thread("Server Shutdown Thread") {
                 @Override
                 public void run() {
                     dedicatedServer.halt(true);
+                    org.apache.logging.log4j.LogManager.shutdown(); // we're manually managing the logging shutdown on the server. Make sure we do it here at the end.
                 }
             };
             shutdownThread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
             Runtime.getRuntime().addShutdownHook(shutdownThread);
+            */ // CraftBukkit end
         } catch (Throwable var40) {
             LOGGER.error(LogUtils.FATAL_MARKER, "Failed to start the minecraft server", var40);
         }
@@ -251,6 +_,9 @@
             dimensions = properties.createDimensions(context.datapackWorldgen());
         }
 
+        // Neo: Do a write-read-cycle to inject modded dimensions on first start of a dedicated server into its generated world dimensions list.
+        var registryOps = net.minecraft.resources.RegistryOps.create(net.minecraft.nbt.NbtOps.INSTANCE, context.datapackWorldgen());
+        dimensions = WorldDimensions.CODEC.encoder().encodeStart(registryOps, dimensions).flatMap((writtenPayloadWithModdedDimensions) -> WorldDimensions.CODEC.decoder().parse(registryOps, writtenPayloadWithModdedDimensions)).resultOrPartial(LOGGER::error).orElse(dimensions);
         WorldDimensions.Complete finalDimensions = dimensions.bake(datapackDimensions);
         Lifecycle lifecycle = finalDimensions.lifecycle().add(context.datapackWorldgen().allRegistriesLifecycle());
         PrimaryLevelData primaryLevelData = new PrimaryLevelData(createLevelSettings, finalDimensions.specialWorldProperty(), lifecycle);
@@ -295,7 +_,7 @@
         RegistryAccess registryAccess,
         boolean recreateRegionFiles
     ) {
-        LOGGER.info("Forcing world upgrade!");
+        LOGGER.info("Forcing world upgrade! {}", storageSource.getLevelId()); // CraftBukkit
 
         try (WorldUpgrader upgrader = new WorldUpgrader(storageSource, fixerUpper, registryAccess, eraseCache, recreateRegionFiles)) {
             Component lastStatus = null;
