--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -10,6 +_,8 @@
 import com.mojang.jtracy.DiscontinuousFrame;
 import com.mojang.jtracy.TracyClient;
 import com.mojang.logging.LogUtils;
+import com.mojang.serialization.Dynamic;
+import com.mojang.serialization.Lifecycle;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ObjectArraySet;
 import java.io.File;
@@ -45,6 +_,7 @@
 import java.util.function.Function;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
+import joptsimple.OptionSet;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportType;
@@ -63,6 +_,8 @@
 import net.minecraft.core.registries.Registries;
 import net.minecraft.data.worldgen.features.MiscOverworldFeatures;
 import net.minecraft.gametest.framework.GameTestTicker;
+import net.minecraft.nbt.NbtException;
+import net.minecraft.nbt.ReportedNbtException;
 import net.minecraft.nbt.Tag;
 import net.minecraft.network.PacketProcessor;
 import net.minecraft.network.chat.ChatDecorator;
@@ -78,6 +_,8 @@
 import net.minecraft.resources.Identifier;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.bossevents.CustomBossEvents;
+import net.minecraft.server.dedicated.DedicatedServer;
+import net.minecraft.server.dedicated.DedicatedServerProperties;
 import net.minecraft.server.level.ChunkLoadCounter;
 import net.minecraft.server.level.ChunkMap;
 import net.minecraft.server.level.DemoMode;
@@ -117,6 +_,7 @@
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.TimeUtil;
 import net.minecraft.util.Util;
+import net.minecraft.util.datafix.DataFixers;
 import net.minecraft.util.debug.ServerDebugSubscribers;
 import net.minecraft.util.debugchart.SampleLogger;
 import net.minecraft.util.debugchart.TpsDebugDimensions;
@@ -135,6 +_,7 @@
 import net.minecraft.util.profiling.metrics.profiling.ServerMetricsSamplersProvider;
 import net.minecraft.util.profiling.metrics.storage.MetricsPersister;
 import net.minecraft.util.thread.ReentrantBlockableEventLoop;
+import net.minecraft.util.worldupdate.UpgradeProgress;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.RandomSequences;
 import net.minecraft.world.Stopwatches;
@@ -163,6 +_,7 @@
 import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.level.chunk.storage.ChunkIOErrorReporter;
 import net.minecraft.world.level.chunk.storage.RegionStorageInfo;
+import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.gamerules.GameRule;
 import net.minecraft.world.level.gamerules.GameRuleMap;
@@ -171,23 +_,35 @@
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.levelgen.PatrolSpawner;
 import net.minecraft.world.level.levelgen.PhantomSpawner;
+import net.minecraft.world.level.levelgen.WorldDimensions;
 import net.minecraft.world.level.levelgen.WorldGenSettings;
 import net.minecraft.world.level.levelgen.WorldOptions;
+import net.minecraft.world.level.levelgen.presets.WorldPresets;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
 import net.minecraft.world.level.saveddata.WeatherData;
 import net.minecraft.world.level.storage.CommandStorage;
 import net.minecraft.world.level.storage.DerivedLevelData;
 import net.minecraft.world.level.storage.LevelData;
+import net.minecraft.world.level.storage.LevelDataAndDimensions;
 import net.minecraft.world.level.storage.LevelResource;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import net.minecraft.world.level.storage.LevelSummary;
 import net.minecraft.world.level.storage.PlayerDataStorage;
+import net.minecraft.world.level.storage.PrimaryLevelData;
 import net.minecraft.world.level.storage.SavedDataStorage;
 import net.minecraft.world.level.storage.ServerLevelData;
 import net.minecraft.world.level.storage.WorldData;
 import net.minecraft.world.level.timers.TimerQueue;
+import net.minecraft.world.level.validation.ContentValidationException;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.ScoreboardSaveData;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.CraftRegistry;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.event.server.ServerLoadEvent;
+import org.jline.terminal.Terminal;
+import org.jline.terminal.TerminalBuilder;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -217,9 +_,9 @@
     public static final NameAndId ANONYMOUS_PLAYER_PROFILE = new NameAndId(Util.NIL_UUID, "Anonymous Player");
     public LevelStorageSource.LevelStorageAccess storageSource;
     public final PlayerDataStorage playerDataStorage;
-    private final SavedDataStorage savedDataStorage;
+    //private final SavedDataStorage savedDataStorage; // CraftBukkit - per world
     private final List<Runnable> tickables = Lists.newArrayList();
-    private final GameRules gameRules;
+    //private final GameRules gameRules; // CraftBukkit - per world
     private MetricsRecorder metricsRecorder = InactiveMetricsRecorder.INSTANCE;
     private Consumer<ProfileResults> onMetricsRecordingStopped = results -> this.stopRecordingMetrics();
     private Consumer<Path> onMetricsRecordingFinished = ignored -> {};
@@ -266,13 +_,13 @@
     private long delayedTasksMaxNextTickTimeNanos;
     private boolean mayHaveDelayedTasks;
     private final PackRepository packRepository;
-    private final WorldGenSettings worldGenSettings;
+    // private final WorldGenSettings worldGenSettings; // CraftBukkit - per world
     private final ServerScoreboard scoreboard = new ServerScoreboard(this);
     private @Nullable Stopwatches stopwatches;
     private @Nullable CommandStorage commandStorage;
-    private final CustomBossEvents customBossEvents;
-    private final RandomSequences randomSequences;
-    private final WeatherData weatherData;
+    // private final CustomBossEvents customBossEvents; // CraftBukkit - per world
+    // private final RandomSequences randomSequences; // CraftBukkit - per world
+    // private final WeatherData weatherData; // CraftBukkit - per world
     private final ServerFunctionManager functionManager;
     private boolean enforceWhitelist;
     private boolean usingWhitelist;
@@ -292,12 +_,25 @@
     private final SuppressedExceptionCollector suppressedExceptions = new SuppressedExceptionCollector();
     private final DiscontinuousFrame tickFrame;
     private final PacketProcessor packetProcessor;
-    private final TimerQueue<MinecraftServer> scheduledEvents;
-    private final ServerClockManager clockManager;
+    // private final TimerQueue<MinecraftServer> scheduledEvents; // CraftBukkit - per world
+    // private final ServerClockManager clockManager; // CraftBukkit - per world
+
+    // CraftBukkit start
+    public final WorldLoader.DataLoadContext worldLoader;
+    public org.bukkit.craftbukkit.CraftServer server;
+    public OptionSet options;
+    public org.bukkit.command.ConsoleCommandSender console;
+    public Terminal terminal;
+    public static int currentTick = (int) (System.currentTimeMillis() / 50);
+    public java.util.Queue<Runnable> processQueue = new java.util.concurrent.ConcurrentLinkedQueue<Runnable>();
+    public int autosavePeriod;
+    public Commands vanillaCommandDispatcher;
+    private boolean forceTicks;
+    // CraftBukkit end
 
     public static <S extends MinecraftServer> S spin(Function<Thread, S> factory) {
         AtomicReference<S> serverReference = new AtomicReference<>();
-        Thread thread = new Thread(() -> serverReference.get().runServer(), "Server thread");
+        Thread thread = new Thread(net.neoforged.fml.util.thread.SidedThreadGroups.SERVER, () -> serverReference.get().runServer(), "Server thread");
         thread.setUncaughtExceptionHandler((t, e) -> LOGGER.error("Uncaught exception in server thread", e));
         if (Runtime.getRuntime().availableProcessors() > 4) {
             thread.setPriority(8);
@@ -310,6 +_,7 @@
     }
 
     public MinecraftServer(
+        OptionSet options, WorldLoader.DataLoadContext worldLoader,
         Thread serverThread,
         LevelStorageSource.LevelStorageAccess storageSource,
         PackRepository packRepository,
@@ -326,10 +_,10 @@
         if (!this.registries.compositeAccess().lookupOrThrow(Registries.LEVEL_STEM).containsKey(LevelStem.OVERWORLD)) {
             throw new IllegalStateException("Missing Overworld dimension data");
         } else {
-            this.savedDataStorage = new SavedDataStorage(storageSource.getLevelPath(LevelResource.DATA), fixerUpper, this.registries.compositeAccess());
+            //this.savedDataStorage = new SavedDataStorage(storageSource.getLevelPath(LevelResource.DATA), fixerUpper, this.registries.compositeAccess());
             this.worldData = worldStem.worldDataAndGenSettings().data();
-            this.worldGenSettings = worldStem.worldDataAndGenSettings().genSettings();
-            this.savedDataStorage.set(WorldGenSettings.TYPE, this.worldGenSettings);
+            //this.worldGenSettings = worldStem.worldDataAndGenSettings().genSettings();
+            //this.savedDataStorage.set(WorldGenSettings.TYPE, this.worldGenSettings);
             this.proxy = proxy;
             this.packRepository = packRepository;
             this.resources = new MinecraftServer.ReloadableResources(worldStem.resourceManager(), worldStem.dataPackResources());
@@ -339,10 +_,10 @@
             this.levelLoadListener = levelLoadListener;
             this.storageSource = storageSource;
             this.playerDataStorage = storageSource.createPlayerStorage();
-            this.randomSequences = this.savedDataStorage.computeIfAbsent(RandomSequences.TYPE);
-            this.weatherData = this.getDataStorage().computeIfAbsent(WeatherData.TYPE);
-            this.gameRules = new GameRules(this.worldData.enabledFeatures(), this.savedDataStorage.computeIfAbsent(GameRuleMap.TYPE));
-            gameRules.ifPresent(g -> this.gameRules.setAll(g, null));
+            //this.randomSequences = this.savedDataStorage.computeIfAbsent(RandomSequences.TYPE);
+            //this.weatherData = this.getDataStorage().computeIfAbsent(WeatherData.TYPE);
+            //this.gameRules = new GameRules(this.worldData.enabledFeatures(), this.savedDataStorage.computeIfAbsent(GameRuleMap.TYPE));
+            //gameRules.ifPresent(g -> this.gameRules.setAll(g, null));
             this.fixerUpper = fixerUpper;
             this.functionManager = new ServerFunctionManager(this, this.resources.managers.getFunctionLibrary());
             HolderGetter<Block> blockLookup = this.registries
@@ -352,18 +_,46 @@
             this.structureTemplateManager = new StructureTemplateManager(worldStem.resourceManager(), storageSource, fixerUpper, blockLookup);
             this.serverThread = serverThread;
             this.executor = Util.backgroundExecutor();
-            this.potionBrewing = PotionBrewing.bootstrap(this.worldData.enabledFeatures());
+            this.potionBrewing = PotionBrewing.bootstrap(this.worldData.enabledFeatures(), this.registryAccess());
             this.resources.managers.getRecipeManager().finalizeRecipeLoading(this.worldData.enabledFeatures());
-            this.fuelValues = FuelValues.vanillaBurnTimes(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+            this.fuelValues = net.neoforged.neoforge.common.DataMapHooks.populateFuelValues(this.registries.compositeAccess(), this.worldData.enabledFeatures());
             this.tickFrame = TracyClient.createDiscontinuousFrame("Server Tick");
             this.notificationManager = new NotificationManager();
             this.serverActivityMonitor = new ServerActivityMonitor(this.notificationManager, 30);
             this.packetProcessor = new PacketProcessor(serverThread);
+            // CraftBukkit start - per world
+            /*
             this.clockManager = this.getDataStorage().computeIfAbsent(ServerClockManager.TYPE);
             this.clockManager.init(this);
             this.customBossEvents = this.savedDataStorage.computeIfAbsent(CustomBossEvents.TYPE);
             this.scheduledEvents = this.savedDataStorage.computeIfAbsent(TimerQueue.TYPE);
-        }
+            */
+            // CraftBukkit end
+        }
+        // CraftBukkit start
+        this.options = options;
+        this.worldLoader = worldLoader;
+        this.vanillaCommandDispatcher = worldStem.dataPackResources().commands; // CraftBukkit
+        // Try to see if we're actually running in a terminal, disable jline if not
+        if (System.console() == null && System.getProperty(TerminalBuilder.PROP_PROVIDER) == null) {
+            System.setProperty(TerminalBuilder.PROP_PROVIDER, Terminal.TYPE_DUMB);
+            org.bukkit.craftbukkit.Main.useJline = false;
+        }
+
+        try {
+            terminal = TerminalBuilder.builder().dumb(false).build();
+        } catch (Throwable e) {
+            try {
+                // Try again with jline disabled for Windows users without C++ 2008 Redistributable
+                System.setProperty(TerminalBuilder.PROP_PROVIDER, Terminal.TYPE_DUMB);
+                org.bukkit.craftbukkit.Main.useJline = false;
+                terminal = TerminalBuilder.builder().build();
+            } catch (IOException ex) {
+                LOGGER.warn((String) null, ex);
+            }
+        }
+        Runtime.getRuntime().addShutdownHook(new org.bukkit.craftbukkit.util.ServerShutdownThread(this));
+        // CraftBukkit end
     }
 
     protected abstract boolean initServer() throws IOException;
@@ -400,15 +_,12 @@
         };
     }
 
-    protected void loadLevel() {
+    protected void loadLevel(String s) { // CraftBukkit
         boolean startedWorldLoadProfiling = !JvmProfiler.INSTANCE.isRunning()
             && SharedConstants.DEBUG_JFR_PROFILING_ENABLE_LEVEL_LOADING
             && JvmProfiler.INSTANCE.start(Environment.from(this));
         ProfiledDuration profiledDuration = JvmProfiler.INSTANCE.onWorldLoadedStarted();
-        this.worldData.setModdedInfo(this.getServerModName(), this.getModdedStatus().shouldReportAsModified());
-        this.createLevels();
-        this.forceDifficulty();
-        this.prepareLevels();
+        this.loadWorld0(s); // CraftBukkit
         if (profiledDuration != null) {
             profiledDuration.finish(true);
         }
@@ -425,65 +_,243 @@
     protected void forceDifficulty() {
     }
 
-    protected void createLevels() {
-        ServerLevelData levelData = this.worldData.overworldData();
-        boolean isDebug = this.worldData.isDebugWorld();
-        Registry<LevelStem> dimensions = this.registries.compositeAccess().lookupOrThrow(Registries.LEVEL_STEM);
-        WorldOptions worldOptions = this.worldGenSettings.options();
-        long seed = worldOptions.seed();
-        long biomeZoomSeed = BiomeManager.obfuscateSeed(seed);
-        List<CustomSpawner> overworldCustomSpawners = ImmutableList.of(
-            new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(this.savedDataStorage)
-        );
-        LevelStem overworldData = dimensions.getValue(LevelStem.OVERWORLD);
-        ServerLevel overworld = new ServerLevel(
-            this, this.executor, this.storageSource, levelData, Level.OVERWORLD, overworldData, isDebug, biomeZoomSeed, overworldCustomSpawners, true
-        );
-        this.levels.put(Level.OVERWORLD, overworld);
-        this.scoreboard.load(this.savedDataStorage.computeIfAbsent(ScoreboardSaveData.TYPE).getData());
-        this.commandStorage = new CommandStorage(this.savedDataStorage);
-        this.stopwatches = this.savedDataStorage.computeIfAbsent(Stopwatches.TYPE);
-        if (!levelData.isInitialized()) {
+    // CraftBukkit start
+    private void loadWorld0(String s) {
+        LevelStorageSource.LevelStorageAccess worldSession = this.storageSource;
+
+        RegistryAccess.Frozen registryaccess_frozen = this.registries.compositeAccess();
+        Registry<LevelStem> dimensions = registryaccess_frozen.lookupOrThrow(Registries.LEVEL_STEM);
+        for (LevelStem worldDimension : dimensions) {
+            ResourceKey<LevelStem> dimensionKey = dimensions.getResourceKey(worldDimension).get();
+
+            ServerLevel world;
+            int dimension = 0;
+
+            if (dimensionKey == LevelStem.NETHER) {
+                if (server.getAllowNether()) {
+                    dimension = -1;
+                } else {
+                    continue;
+                }
+            } else if (dimensionKey == LevelStem.END) {
+                if (server.getAllowEnd()) {
+                    dimension = 1;
+                } else {
+                    continue;
+                }
+            } else if (dimensionKey != LevelStem.OVERWORLD) {
+                dimension = -999;
+            }
+
+            String worldType = (dimension == -999) ? dimensionKey.identifier().getNamespace() + "_" + dimensionKey.identifier().getPath() : org.bukkit.World.Environment.getEnvironment(dimension).toString().toLowerCase(Locale.ROOT);
+            String name = (dimensionKey == LevelStem.OVERWORLD) ? s : s + "_" + worldType;
+            if (dimension != 0) {
+                File newWorldFolder = new File(name);
+                File newWorld = LevelStorageSource.getPre261StorageFolder(newWorldFolder.toPath(), dimensionKey).toFile();
+                File oldWorld = LevelStorageSource.getPre261StorageFolder(new File(s).toPath(), dimensionKey).toFile();
+                if (!oldWorld.isDirectory()) {
+                    newWorld = DimensionType.getStorageFolder(Registries.levelStemToLevel(dimensionKey), new File(name).toPath()).toFile();
+                    oldWorld = DimensionType.getStorageFolder(Registries.levelStemToLevel(dimensionKey), new File(s).toPath()).toFile();
+                }
+                File oldLevelDat = new File(new File(s), "level.dat"); // The data folders exist on first run as they are created in the PersistentCollection constructor above, but the level.dat won't
+
+                if (!newWorldFolder.isDirectory() && !newWorld.isDirectory() && oldWorld.isDirectory() && oldLevelDat.isFile()) {
+                    MinecraftServer.LOGGER.info("---- Migration of old " + worldType + " folder required ----");
+                    MinecraftServer.LOGGER.info("Unfortunately due to the way that Minecraft implemented multiworld support in 1.6, Bukkit requires that you move your " + worldType + " folder to a new location in order to operate correctly.");
+                    MinecraftServer.LOGGER.info("We will move this folder for you, but it will mean that you need to move it back should you wish to stop using Bukkit in the future.");
+                    MinecraftServer.LOGGER.info("Attempting to move " + oldWorld + " to " + newWorld + "...");
+
+                    if (newWorld.exists()) {
+                        MinecraftServer.LOGGER.warn("A file or folder already exists at " + newWorld + "!");
+                        MinecraftServer.LOGGER.info("---- Migration of old " + worldType + " folder failed ----");
+                    } else if (newWorld.getParentFile().mkdirs()) {
+                        if (oldWorld.renameTo(newWorld)) {
+                            MinecraftServer.LOGGER.info("Success! To restore " + worldType + " in the future, simply move " + newWorld + " to " + oldWorld);
+                            // Migrate world data too.
+                            try {
+                                com.google.common.io.Files.copy(oldLevelDat, new File(new File(name), "level.dat"));
+                                org.apache.commons.io.FileUtils.copyDirectory(new File(new File(s), "data"), new File(new File(name), "data"));
+                            } catch (IOException exception) {
+                                MinecraftServer.LOGGER.warn("Unable to migrate world data.");
+                            }
+                            MinecraftServer.LOGGER.info("---- Migration of old " + worldType + " folder complete ----");
+                        } else {
+                            MinecraftServer.LOGGER.warn("Could not move folder " + oldWorld + " to " + newWorld + "!");
+                            MinecraftServer.LOGGER.info("---- Migration of old " + worldType + " folder failed ----");
+                        }
+                    } else {
+                        MinecraftServer.LOGGER.warn("Could not create path for " + newWorld + "!");
+                        MinecraftServer.LOGGER.info("---- Migration of old " + worldType + " folder failed ----");
+                    }
+                }
+
+                try {
+                    worldSession = LevelStorageSource.createDefault(server.getWorldContainer().toPath()).validateAndCreateAccess(name, dimensionKey);
+                } catch (IOException | ContentValidationException ex) {
+                    throw new RuntimeException(ex);
+                }
+            }
+
+            Dynamic<?> dynamic;
+            if (worldSession.hasWorldData()) {
+                Dynamic<?> dynamic1;
+
+                try {
+                    dynamic1 = worldSession.getUnfixedDataTagWithFallback();
+                } catch (NbtException | ReportedNbtException | IOException ioexception) {
+                    MinecraftServer.LOGGER.error("Failed to load world data. World files may be corrupted. Shutting down.", ioexception);
+                    return;
+                }
+
+                LevelSummary levelsummary = worldSession.fixAndGetSummaryFromTag(dynamic1);
+
+                if (levelsummary.requiresManualConversion()) {
+                    MinecraftServer.LOGGER.info("This world must be opened in an older version (like 1.6.4) to be safely converted");
+                    return;
+                }
+
+                if (!levelsummary.isCompatible()) {
+                    MinecraftServer.LOGGER.info("This world was created by an incompatible version.");
+                    return;
+                }
+
+                dynamic = DataFixers.getFileFixer().fix(worldSession, dynamic1, new UpgradeProgress());
+            } else {
+                dynamic = null;
+            }
+
+            org.bukkit.generator.ChunkGenerator gen = this.server.getGenerator(name);
+            org.bukkit.generator.BiomeProvider biomeProvider = this.server.getBiomeProvider(name);
+
+            LevelDataAndDimensions.WorldDataAndGenSettings dataAndSettings;
+            WorldLoader.DataLoadContext dataloadcontext = this.worldLoader;
+            Registry<LevelStem> registry = dataloadcontext.datapackDimensions().lookupOrThrow(Registries.LEVEL_STEM);
+            if (dynamic != null) {
+                LevelDataAndDimensions leveldataanddimensions = LevelStorageSource.getLevelDataAndDimensions(worldSession, dynamic, dataloadcontext.dataConfiguration(), registry, dataloadcontext.datapackWorldgen());
+
+                dataAndSettings = leveldataanddimensions.worldDataAndGenSettings();
+            } else {
+                LevelSettings levelsettings;
+                WorldOptions worldoptions;
+                WorldDimensions worlddimensions;
+
+                if (this.isDemo()) {
+                    levelsettings = MinecraftServer.DEMO_SETTINGS;
+                    worldoptions = WorldOptions.DEMO_OPTIONS;
+                    worlddimensions = WorldPresets.createNormalWorldDimensions(dataloadcontext.datapackWorldgen());
+                } else {
+                    DedicatedServerProperties dedicatedserverproperties = ((DedicatedServer) this).getProperties();
+
+                    levelsettings = new LevelSettings(dedicatedserverproperties.levelName, dedicatedserverproperties.gameMode.get(), new LevelSettings.DifficultySettings(dedicatedserverproperties.difficulty.get(), dedicatedserverproperties.hardcore, false), false, dataloadcontext.dataConfiguration());
+                    worldoptions = options.has("bonusChest") ? dedicatedserverproperties.worldOptions.withBonusChest(true) : dedicatedserverproperties.worldOptions;
+                    worlddimensions = dedicatedserverproperties.createDimensions(dataloadcontext.datapackWorldgen());
+                }
+
+                WorldDimensions.Complete worlddimensions_complete = worlddimensions.bake(registry);
+                Lifecycle lifecycle = worlddimensions_complete.lifecycle().add(dataloadcontext.datapackWorldgen().allRegistriesLifecycle());
+
+                PrimaryLevelData primaryleveldata = new PrimaryLevelData(levelsettings, worlddimensions_complete.specialWorldProperty(), lifecycle);
+                dataAndSettings = new LevelDataAndDimensions.WorldDataAndGenSettings(primaryleveldata, new WorldGenSettings(worldoptions, worlddimensions));
+            }
+            PrimaryLevelData worlddata = (PrimaryLevelData) dataAndSettings.data();
+            worlddata.checkName(name); // CraftBukkit - Migration did not rewrite the level.dat; This forces 1.8 to take the last loaded world as respawn (in this case the end)
+            if (options.has("forceUpgrade")) {
+                net.minecraft.server.Main.forceUpgrade(worldSession, DataFixers.getDataFixer(), options.has("eraseCache"), () -> {
+                    return true;
+                }, registryaccess_frozen, options.has("recreateRegionFiles"));
+            }
+            SavedDataStorage dimensiondatastorage = new SavedDataStorage(worldSession.getLevelPath(LevelResource.DATA), this.getFixerUpper(), this.registryAccess());
+
+            ServerLevelData serverleveldata = worlddata.overworldData();
+            boolean flag = worlddata.isDebugWorld();
+            WorldOptions worldoptions = dataAndSettings.genSettings().options();
+            long i = worldoptions.seed();
+            long j = BiomeManager.obfuscateSeed(i);
+            List<CustomSpawner> list = ImmutableList.of(new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(dimensiondatastorage));
+            LevelStem levelstem = (LevelStem) dimensions.getValue(dimensionKey);
+
+            org.bukkit.generator.WorldInfo worldInfo = new org.bukkit.craftbukkit.generator.CraftWorldInfo(serverleveldata, worldoptions, worldSession, org.bukkit.World.Environment.getEnvironment(dimension), levelstem.type().value());
+            if (biomeProvider == null && gen != null) {
+                biomeProvider = gen.getDefaultBiomeProvider(worldInfo);
+            }
+
+            ResourceKey<Level> worldKey = ResourceKey.create(Registries.DIMENSION, dimensionKey.identifier());
+
+            if (dimensionKey == LevelStem.OVERWORLD) {
+                this.worldData = worlddata;
+                this.worldData.setGameType(((DedicatedServer) this).getProperties().gameMode.get()); // From DedicatedServer.init
+
+                world = new ServerLevel(this, this.executor, worldSession, worlddata, worldKey, levelstem, flag, j, list, true, dimensiondatastorage, dataAndSettings.genSettings(), org.bukkit.World.Environment.getEnvironment(dimension), gen, biomeProvider);
+                this.scoreboard.load(((ScoreboardSaveData) dimensiondatastorage.computeIfAbsent(ScoreboardSaveData.TYPE)).getData());
+                this.server.scoreboardManager = new org.bukkit.craftbukkit.scoreboard.CraftScoreboardManager(this, world.getScoreboard());
+                this.commandStorage = new CommandStorage(dimensiondatastorage);
+                this.stopwatches = (Stopwatches) dimensiondatastorage.computeIfAbsent(Stopwatches.TYPE);
+            } else {
+                world = new ServerLevel(this, this.executor, worldSession, worlddata, worldKey, levelstem, flag, j, ImmutableList.of(), true, dimensiondatastorage, dataAndSettings.genSettings(), org.bukkit.World.Environment.getEnvironment(dimension), gen, biomeProvider);
+            }
+
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Load(world));
+            worlddata.setModdedInfo(this.getServerModName(), this.getModdedStatus().shouldReportAsModified());
+            this.initWorld(world, worlddata, worldData, worldoptions);
+
+            this.addLevel(world);
+            this.getPlayerList().addWorldborderListener(world);
+        }
+        this.forceDifficulty();
+        for (ServerLevel serverlevel : this.getAllLevels()) {
+            this.prepareLevels(serverlevel);
+            serverlevel.entityManager.tick(); // SPIGOT-6526: Load pending entities so they are available to the API
+            this.server.getPluginManager().callEvent(new org.bukkit.event.world.WorldLoadEvent(serverlevel.getWorld()));
+        }
+
+        this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
+        this.server.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.STARTUP));
+        this.connection.acceptConnections();
+    }
+
+    public void initWorld(ServerLevel serverlevel, ServerLevelData serverleveldata, WorldData saveData, WorldOptions worldoptions) {
+        boolean flag = saveData.isDebugWorld();
+        // CraftBukkit start
+        if (serverlevel.generator != null) {
+            serverlevel.getWorld().getPopulators().addAll(serverlevel.generator.getDefaultPopulators(serverlevel.getWorld()));
+        }
+        initWorldBorder(serverlevel); // CraftBukkit - move up so that WorldBorder is set during WorldInitEvent
+        this.server.getPluginManager().callEvent(new org.bukkit.event.world.WorldInitEvent(serverlevel.getWorld())); // CraftBukkit - SPIGOT-5569: Call WorldInitEvent before any chunks are generated
+
+        if (!serverleveldata.isInitialized()) {
             try {
-                setInitialSpawn(overworld, levelData, worldOptions.generateBonusChest(), isDebug, this.levelLoadListener);
-                levelData.setInitialized(true);
-                if (isDebug) {
-                    this.setupDebugLevel(this.worldData);
+                setInitialSpawn(serverlevel, serverleveldata, worldoptions.generateBonusChest(), flag, this.levelLoadListener);
+                serverleveldata.setInitialized(true);
+                if (flag) {
+                    this.setupDebugLevel(saveData, serverlevel); // CraftBukkit
                 }
-            } catch (Throwable var20) {
-                CrashReport report = CrashReport.forThrowable(var20, "Exception initializing level");
+            } catch (Throwable throwable) {
+                CrashReport crashreport = CrashReport.forThrowable(throwable, "Exception initializing level");
 
                 try {
-                    overworld.fillReportDetails(report);
-                } catch (Throwable var19) {
+                    serverlevel.fillReportDetails(crashreport);
+                } catch (Throwable throwable1) {
+                    ;
                 }
 
-                throw new ReportedException(report);
-            }
-
-            levelData.setInitialized(true);
-        }
-
-        GlobalPos focusPos = this.selectLevelLoadFocusPos();
-        this.levelLoadListener.updateFocus(focusPos.dimension(), ChunkPos.containing(focusPos.pos()));
-
-        for (Entry<ResourceKey<LevelStem>, LevelStem> entry : dimensions.entrySet()) {
-            ResourceKey<LevelStem> name = entry.getKey();
-            ServerLevel level;
-            if (name != LevelStem.OVERWORLD) {
-                ResourceKey<Level> dimension = ResourceKey.create(Registries.DIMENSION, name.identifier());
-                DerivedLevelData derivedLevelData = new DerivedLevelData(this.worldData, levelData);
-                level = new ServerLevel(
-                    this, this.executor, this.storageSource, derivedLevelData, dimension, entry.getValue(), isDebug, biomeZoomSeed, ImmutableList.of(), false
-                );
-                this.levels.put(dimension, level);
-            } else {
-                level = overworld;
-            }
-
-            level.getWorldBorder().setAbsoluteMaxSize(this.getAbsoluteMaxWorldSize());
-            this.getPlayerList().addWorldborderListener(level);
-        }
+                throw new ReportedException(crashreport);
+            }
+
+            serverleveldata.setInitialized(true);
+        }
+
+        GlobalPos globalpos = new GlobalPos(serverlevel.dimension(), serverlevel.getRespawnData().pos());
+
+        this.levelLoadListener.updateFocus(globalpos.dimension(), ChunkPos.containing(globalpos.pos()));
+    }
+
+    private void initWorldBorder(ServerLevel serverlevel1) {
+        {
+            // CraftBukkit end
+            serverlevel1.getWorldBorder().setAbsoluteMaxSize(this.getAbsoluteMaxWorldSize());
+            this.getPlayerList().addWorldborderListener(serverlevel1);
+        }
+
     }
 
     private static void setInitialSpawn(
@@ -495,6 +_,7 @@
             levelData.setSpawn(LevelData.RespawnData.of(level.dimension(), BlockPos.ZERO.above(80), 0.0F, 0.0F));
         } else {
             ServerChunkCache chunkSource = level.getChunkSource();
+            if (net.neoforged.neoforge.event.EventHooks.onCreateWorldSpawn(level, levelData)) return;
             ChunkPos spawnChunk = ChunkPos.containing(chunkSource.randomState().sampler().findSpawnPosition());
             levelLoadListener.start(LevelLoadListener.Stage.PREPARE_GLOBAL_SPAWN, 0);
             levelLoadListener.updateFocus(level.dimension(), spawnChunk);
@@ -540,37 +_,41 @@
         }
     }
 
-    private void setupDebugLevel(WorldData worldData) {
+    private void setupDebugLevel(WorldData worldData, ServerLevel serverlevel) { // CraftBukkit - added level
         worldData.setDifficulty(Difficulty.PEACEFUL);
         worldData.setDifficultyLocked(true);
         ServerLevelData levelData = worldData.overworldData();
-        this.getGameRules().set(GameRules.ADVANCE_WEATHER, false, this);
-        this.clockManager.moveToTimeMarker(this.registryAccess().getOrThrow(WorldClocks.OVERWORLD), ClockTimeMarkers.NOON);
+        this.getGameRules().set(GameRules.ADVANCE_WEATHER, false, serverlevel); // CraftBukkit
+        serverlevel.clockManager().moveToTimeMarker(this.registryAccess().getOrThrow(WorldClocks.OVERWORLD), ClockTimeMarkers.NOON); // CraftBukkit
         levelData.setGameType(GameType.SPECTATOR);
     }
 
-    private void prepareLevels() {
+    public void prepareLevels(ServerLevel serverlevel) {
         ChunkLoadCounter chunkLoadCounter = new ChunkLoadCounter();
 
-        for (ServerLevel level : this.levels.values()) {
-            chunkLoadCounter.track(level, () -> {
-                TicketStorage savedTickets = level.getDataStorage().get(TicketStorage.TYPE);
+            chunkLoadCounter.track(serverlevel, () -> {
+                TicketStorage savedTickets = serverlevel.getDataStorage().get(TicketStorage.TYPE);
                 if (savedTickets != null) {
                     savedTickets.activateAllDeactivatedTickets();
+                    net.neoforged.neoforge.common.world.chunk.ForcedChunkManager.activateAllDeactivatedTickets(serverlevel, savedTickets);
                 }
             });
-        }
 
         this.levelLoadListener.start(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS, chunkLoadCounter.totalChunks());
 
         do {
             this.levelLoadListener.update(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS, chunkLoadCounter.readyChunks(), chunkLoadCounter.totalChunks());
-            this.nextTickTimeNanos = Util.getNanos() + PREPARE_LEVELS_DEFAULT_DELAY_NANOS;
-            this.waitUntilNextTick();
+            // CraftBukkit start
+            // this.nextTickTimeNanos = Util.getNanos() + PREPARE_LEVELS_DEFAULT_DELAY_NANOS;
+            this.executeModerately();
+            // CraftBukkit end
         } while (chunkLoadCounter.pendingChunks() > 0);
 
         this.levelLoadListener.finish(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS);
-        this.updateMobSpawningFlags();
+        // CraftBukkit start
+        // this.updateMobSpawningFlags();
+        this.forceTicks = false;
+        // CraftBukkit end
         this.updateEffectiveRespawnData();
     }
 
@@ -604,7 +_,8 @@
             level.save(null, flush, SharedConstants.DEBUG_DONT_SAVE_WORLD || level.noSave && !force);
             result = true;
         }
-
+        // CraftBukkit start - moved to ServerLevel#save
+        /*
         GameProfile singleplayerProfile = this.getSingleplayerProfile();
         this.storageSource.saveDataTag(this.worldData, singleplayerProfile == null ? null : singleplayerProfile.id());
         if (flush) {
@@ -612,6 +_,8 @@
         } else {
             this.savedDataStorage.scheduleSave();
         }
+        */
+        // CraftBukkit end
 
         if (flush) {
             for (ServerLevel level : this.getAllLevels()) {
@@ -645,13 +_,34 @@
         this.stopServer();
     }
 
+    // CraftBukkit start
+    private boolean hasStopped = false;
+    private final Object stopLock = new Object();
+    public final boolean hasStopped() {
+        synchronized (stopLock) {
+            return hasStopped;
+        }
+    }
+    // CraftBukkit end
+
     protected void stopServer() {
+        // CraftBukkit start - prevent double stopping on multiple threads
+        synchronized(stopLock) {
+            if (hasStopped) return;
+            hasStopped = true;
+        }
+        // CraftBukkit end
         this.packetProcessor.close();
         if (this.metricsRecorder.isRecording()) {
             this.cancelRecordingMetrics();
         }
 
         LOGGER.info("Stopping server");
+        // CraftBukkit start
+        if (this.server != null) {
+            this.server.disablePlugins();
+        }
+        // CraftBukkit end
         this.getConnection().stop();
         this.isSaving = true;
         if (this.playerList != null) {
@@ -684,6 +_,7 @@
         for (ServerLevel levelx : this.getAllLevels()) {
             if (levelx != null) {
                 try {
+                    net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Unload(levelx));
                     levelx.close();
                 } catch (IOException var5) {
                     LOGGER.error("Exception closing the level", (Throwable)var5);
@@ -692,7 +_,7 @@
         }
 
         this.isSaving = false;
-        this.savedDataStorage.close();
+        // this.savedDataStorage.close(); // CraftBukkit - handled in ServerLevel#close
         this.resources.close();
 
         try {
@@ -731,9 +_,11 @@
                 throw new IllegalStateException("Failed to initialize server");
             }
 
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarted(this);
             this.nextTickTimeNanos = Util.getNanos();
             this.statusIcon = this.loadStatusIcon().orElse(null);
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
 
             while (this.running) {
                 long thisTickNanos;
@@ -787,6 +_,8 @@
                 this.isReady = true;
                 JvmProfiler.INSTANCE.onServerTick(this.smoothedTickTimeMillis);
             }
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStopping(this);
+            net.neoforged.neoforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
         } catch (Throwable var69) {
             LOGGER.error("Encountered an unexpected exception", var69);
             CrashReport report = constructOrExtractCrashReport(var69);
@@ -798,6 +_,7 @@
                 LOGGER.error("We were unable to save this crash report to disk.");
             }
 
+            net.neoforged.neoforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
             this.onServerCrash(report);
         } finally {
             try {
@@ -806,6 +_,7 @@
             } catch (Throwable var64) {
                 LOGGER.error("Exception stopping the server", var64);
             } finally {
+                net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStopped(this);
                 this.onServerExit();
             }
         }
@@ -858,7 +_,14 @@
     }
 
     private boolean haveTime() {
-        return this.runningTask() || Util.getNanos() < (this.mayHaveDelayedTasks ? this.delayedTasksMaxNextTickTimeNanos : this.nextTickTimeNanos);
+        // CraftBukkit start
+        return this.forceTicks || this.runningTask() || Util.getNanos() < (this.mayHaveDelayedTasks ? this.delayedTasksMaxNextTickTimeNanos : this.nextTickTimeNanos);
+    }
+
+    private void executeModerately() {
+        this.runAllTasks();
+        java.util.concurrent.locks.LockSupport.parkNanos("executing tasks", 1000L);
+        // CraftBukkit end
     }
 
     public NotificationManager notificationManager() {
@@ -987,11 +_,13 @@
         }
 
         this.tickCount++;
+        net.neoforged.neoforge.event.EventHooks.fireServerTickPre(haveTime, this);
         this.tickRateManager.tick();
         this.tickChildren(haveTime);
         if (nano - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) {
             this.lastServerStatus = nano;
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
         }
 
         this.ticksUntilAutosave--;
@@ -1009,6 +_,7 @@
         this.smoothedTickTimeMillis = this.smoothedTickTimeMillis * 0.8F + (float)tickTime / (float)TimeUtil.NANOSECONDS_PER_MILLISECOND * 0.19999999F;
         this.logTickMethodTime(nano);
         profiler.pop();
+        net.neoforged.neoforge.event.EventHooks.fireServerTickPost(haveTime, this);
     }
 
     protected void processPacketsAndTick(boolean sprinting) {
@@ -1039,6 +_,16 @@
         }
     }
 
+    private static final com.google.gson.Gson GSON = new com.google.gson.Gson();
+    private String cachedServerStatus; // Neo: cache the server status json in case a client spams requests
+    private void resetStatusCache(ServerStatus status) {
+        this.cachedServerStatus = GSON.toJson(ServerStatus.CODEC.encodeStart(com.mojang.serialization.JsonOps.INSTANCE, status)
+                .result().orElseThrow());
+    }
+    public String getStatusJson() {
+        return cachedServerStatus;
+    }
+
     private int computeNextAutosaveInterval() {
         float ticksPerSecond;
         if (this.tickRateManager.isSprinting()) {
@@ -1070,7 +_,8 @@
             Optional.of(players),
             Optional.of(ServerStatus.Version.current()),
             Optional.ofNullable(this.statusIcon),
-            this.enforceSecureProfile()
+            this.enforceSecureProfile(),
+            true //TODO Neo: Possible build a system which indicates what the status of the modded server is.
         );
     }
 
@@ -1102,7 +_,11 @@
         profiler.pop();
         if (this.tickRateManager.runsNormally()) {
             profiler.push("clocks");
-            this.clockManager.tick();
+            // CraftBukkit start
+            for (ServerLevel serverlevel : this.getAllLevels()) {
+                serverlevel.clockManager().tick();
+            }
+            // CraftBukkit end
             profiler.pop();
         }
 
@@ -1115,9 +_,11 @@
         profiler.push("levels");
         this.updateEffectiveRespawnData();
 
-        for (ServerLevel level : this.getAllLevels()) {
+        for (ServerLevel level : this.getWorldArray()) {
+            long tickStart = Util.getNanos();
             profiler.push(() -> level + " " + level.dimension().identifier());
             profiler.push("tick");
+            net.neoforged.neoforge.event.EventHooks.fireLevelTickPre(level, haveTime);
 
             try {
                 level.tick(haveTime);
@@ -1126,9 +_,11 @@
                 level.fillReportDetails(report);
                 throw new ReportedException(report);
             }
+            net.neoforged.neoforge.event.EventHooks.fireLevelTickPost(level, haveTime);
 
             profiler.pop();
             profiler.pop();
+            perWorldTickTimes.computeIfAbsent(level.dimension(), k -> new long[100])[this.tickCount % 100] = Util.getNanos() - tickStart;
         }
 
         profiler.popPush("connection");
@@ -1137,8 +_,8 @@
         this.playerList.tick();
         profiler.popPush("debugSubscribers");
         this.debugSubscribers.tick();
-        if (this.tickRateManager.runsNormally()) {
             profiler.popPush("gameTests");
+        if (net.neoforged.neoforge.gametest.GameTestHooks.isGametestEnabled() && this.tickRateManager.runsNormally()) {
             GameTestTicker.SINGLETON.tick();
         }
 
@@ -1170,6 +_,9 @@
     }
 
     public void forceGameTimeSynchronization() {
+        if (this.overworld() == null) {
+            return; // Neo: Our ephemeral JUnit server has no overworld
+        }
         ProfilerFiller profiler = Profiler.get();
         profiler.push("timeSync");
         this.playerList.broadcastAll(new ClientboundSetTimePacket(this.overworld().getGameTime(), Map.of()));
@@ -1200,6 +_,22 @@
         return this.levels.get(dimension);
     }
 
+    // CraftBukkit start
+    public void addLevel(ServerLevel level) {
+        Map<ResourceKey<Level>, ServerLevel> oldLevels = this.levels;
+        Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
+        newLevels.put(level.dimension(), level);
+        this.levels = Collections.unmodifiableMap(newLevels);
+    }
+
+    public void removeLevel(ServerLevel level) {
+        Map<ResourceKey<Level>, ServerLevel> oldLevels = this.levels;
+        Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
+        newLevels.remove(level.dimension());
+        this.levels = Collections.unmodifiableMap(newLevels);
+    }
+    // CraftBukkit end
+
     public Set<ResourceKey<Level>> levelKeys() {
         return this.levels.keySet();
     }
@@ -1223,11 +_,11 @@
     }
 
     public String getServerModName() {
-        return "vanilla";
+        return net.neoforged.neoforge.internal.BrandingControl.getServerBranding();
     }
 
     public ServerClockManager clockManager() {
-        return this.clockManager;
+        return this.overworld().clockManager(); // CraftBukkit
     }
 
     public SystemReport fillSystemReport(SystemReport systemReport) {
@@ -1245,7 +_,7 @@
             () -> FeatureFlags.REGISTRY.toNames(this.worldData.enabledFeatures()).stream().map(Identifier::toString).collect(Collectors.joining(", "))
         );
         systemReport.setDetail("World Generation", () -> this.worldData.worldGenSettingsLifecycle().toString());
-        systemReport.setDetail("World Seed", () -> String.valueOf(this.worldGenSettings.options().seed()));
+        systemReport.setDetail("World Seed", (this.overworld() != null) ? String.valueOf(this.getWorldGenSettings().options().seed()) : "");
         systemReport.setDetail("Suppressed Exceptions", this.suppressedExceptions::dump);
         if (this.serverId != null) {
             systemReport.setDetail("Server Id", () -> this.serverId);
@@ -1517,9 +_,9 @@
 
     public CompletableFuture<Void> reloadResources(Collection<String> packsToEnable) {
         CompletableFuture<Void> result = CompletableFuture.<ImmutableList>supplyAsync(
-                () -> packsToEnable.stream()
-                    .map(this.packRepository::getPack)
-                    .filter(Objects::nonNull)
+                () -> this.packRepository
+                    .rebuildSelected(packsToEnable)
+                    .stream()
                     .map(Pack::open)
                     .collect(ImmutableList.toImmutableList()),
                 this
@@ -1558,7 +_,8 @@
                 this.getPlayerList().reloadResources();
                 this.functionManager.replaceLibrary(this.resources.managers.getFunctionLibrary());
                 this.structureTemplateManager.onResourceManagerReload(this.resources.resourceManager);
-                this.fuelValues = FuelValues.vanillaBurnTimes(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+                this.fuelValues = net.neoforged.neoforge.common.DataMapHooks.populateFuelValues(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+                this.getPlayerList().getPlayers().forEach(this.getPlayerList()::sendPlayerPermissionLevel); //Forge: Fix newly added/modified commands not being sent to the client when commands reload.
             }, this);
         if (this.isSameThread()) {
             this.managedBlock(result::isDone);
@@ -1574,8 +_,10 @@
         FeatureFlagSet forcedFeatures = initMode ? FeatureFlagSet.of() : initialDataConfig.enabledFeatures();
         FeatureFlagSet allowedFeatures = initMode ? FeatureFlags.REGISTRY.allFlags() : initialDataConfig.enabledFeatures();
         packRepository.reload();
+        DataPackConfig.DEFAULT.addModPacks(net.neoforged.neoforge.common.CommonHooks.getModDataPacks());
+        dataPackConfig.addModPacks(net.neoforged.neoforge.common.CommonHooks.getModDataPacks());
         if (safeMode) {
-            return configureRepositoryWithSelection(packRepository, List.of("vanilla"), forcedFeatures, false);
+            return configureRepositoryWithSelection(packRepository, net.neoforged.neoforge.common.CommonHooks.getModDataPacksWithVanilla(), forcedFeatures, false);
         } else {
             Set<String> selected = Sets.newLinkedHashSet();
 
@@ -1621,6 +_,8 @@
                 selected.add("vanilla");
             }
 
+            net.neoforged.neoforge.resource.ResourcePackLoader.reorderNewlyDiscoveredPacks(selected, dataPackConfig.getEnabled(), packRepository);
+
             return configureRepositoryWithSelection(packRepository, selected, forcedFeatures, true);
         }
     }
@@ -1742,7 +_,7 @@
     public abstract boolean shouldInformAdmins();
 
     public WorldGenSettings getWorldGenSettings() {
-        return this.worldGenSettings;
+        return this.overworld().getWorldGenSettings(); // CraftBukkit
     }
 
     public RecipeManager getRecipeManager() {
@@ -1770,15 +_,15 @@
     }
 
     public CustomBossEvents getCustomBossEvents() {
-        return this.customBossEvents;
+        return this.overworld().getCustomBossEvents(); // CraftBukkit
     }
 
     public RandomSource getRandomSequence(Identifier key) {
-        return this.randomSequences.get(key, this.worldGenSettings.options().seed());
+        return this.overworld().getRandomSequence(key); // CraftBukkit
     }
 
     public RandomSequences getRandomSequences() {
-        return this.randomSequences;
+        return this.overworld().getRandomSequences(); // CraftBukkit
     }
 
     public void setWeatherParameters(int clearTime, int rainTime, boolean raining, boolean thundering) {
@@ -1791,7 +_,7 @@
     }
 
     public WeatherData getWeatherData() {
-        return this.weatherData;
+        return this.overworld().getWeatherData(); // CraftBukkit
     }
 
     public boolean isEnforceWhitelist() {
@@ -1845,6 +_,31 @@
 
     public abstract boolean isSingleplayerOwner(NameAndId nameAndId);
 
+    private Map<ResourceKey<Level>, long[]> perWorldTickTimes = Maps.newIdentityHashMap();
+    @Nullable
+    public long[] getTickTime(ResourceKey<Level> dim) {
+        return perWorldTickTimes.get(dim);
+    }
+
+    @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+    public synchronized Map<ResourceKey<Level>, ServerLevel> forgeGetWorldMap() {
+        return this.levels;
+    }
+    private int worldArrayMarker = 0;
+    private int worldArrayLast = -1;
+    private ServerLevel[] worldArray;
+    @Deprecated //Forge Internal use Only, use to protect against concurrent modifications in the world tick loop.
+    public synchronized void markWorldsDirty() {
+        worldArrayMarker++;
+    }
+    private ServerLevel[] getWorldArray() {
+        if (worldArrayMarker == worldArrayLast && worldArray != null)
+            return worldArray;
+        worldArray = this.levels.values().stream().toArray(x -> new ServerLevel[x]);
+        worldArrayLast = worldArrayMarker;
+        return worldArray;
+    }
+
     public void dumpServerProperties(Path path) throws IOException {
     }
 
@@ -1944,6 +_,22 @@
         }
     }
 
+    // CraftBukkit start
+    public boolean isDebugging() {
+        return false;
+    }
+
+    @Deprecated
+    public static MinecraftServer getServer() {
+        return (Bukkit.getServer() instanceof CraftServer) ? ((CraftServer) Bukkit.getServer()).getServer() : null;
+    }
+
+    @Deprecated
+    public static RegistryAccess getDefaultRegistryAccess() {
+        return CraftRegistry.getMinecraftRegistry();
+    }
+    // CraftBukkit end
+
     private ProfilerFiller createProfiler() {
         if (this.willStartRecordingMetrics) {
             this.metricsRecorder = ActiveMetricsRecorder.createStarted(
@@ -2009,6 +_,10 @@
         return this.worldData;
     }
 
+    public MinecraftServer.ReloadableResources getServerResources() {
+         return resources;
+    }
+
     public RegistryAccess.Frozen registryAccess() {
         return this.registries.compositeAccess();
     }
@@ -2111,12 +_,12 @@
         return false;
     }
 
-    public <T> void onGameRuleChanged(GameRule<T> rule, T value) {
-        this.notificationManager().onGameRuleChanged(rule, value);
+    public static <T> void onGameRuleChanged(GameRule<T> rule, T value, ServerLevel serverlevel) { // CraftBukkit - per-world
+        serverlevel.getServer().notificationManager().onGameRuleChanged(rule, value);
         if (rule == GameRules.REDUCED_DEBUG_INFO) {
             byte event = (byte)((Boolean)value ? 22 : 23);
 
-            for (ServerPlayer player : this.getPlayerList().getPlayers()) {
+            for (ServerPlayer player : serverlevel.players()) {
                 player.connection.send(new ClientboundEntityEventPacket(player, event));
             }
         } else if (rule == GameRules.LIMITED_CRAFTING || rule == GameRules.IMMEDIATE_RESPAWN) {
@@ -2124,21 +_,26 @@
                 ? ClientboundGameEventPacket.LIMITED_CRAFTING
                 : ClientboundGameEventPacket.IMMEDIATE_RESPAWN;
             ClientboundGameEventPacket packet = new ClientboundGameEventPacket(eventType, (Boolean)value ? 1.0F : 0.0F);
-            this.getPlayerList().getPlayers().forEach(playerx -> playerx.connection.send(packet));
+            serverlevel.players().forEach(playerx -> playerx.connection.send(packet));
         } else if (rule == GameRules.LOCATOR_BAR) {
-            this.getAllLevels().forEach(level -> {
-                ServerWaypointManager waypointManager = level.getWaypointManager();
-                if ((Boolean)value) {
-                    level.players().forEach(waypointManager::updatePlayer);
-                } else {
-                    waypointManager.breakAllConnections();
-                }
-            });
+            ServerWaypointManager serverwaypointmanager = serverlevel.getWaypointManager();
+
+            if ((Boolean) value) {
+                List<ServerPlayer> list = serverlevel.players(); // CraftBukkit - decompile error
+
+                Objects.requireNonNull(serverwaypointmanager);
+                list.forEach(serverwaypointmanager::updatePlayer);
+            } else {
+                serverwaypointmanager.breakAllConnections();
+            }
         } else if (rule == GameRules.SPAWN_MONSTERS) {
-            this.updateMobSpawningFlags();
+            serverlevel.setSpawnSettings(serverlevel.isSpawningMonsters()); // CraftBukkit - per-world
         } else if (rule == GameRules.ADVANCE_TIME) {
-            this.getPlayerList().broadcastAll(this.clockManager().createFullSyncPacket());
+            serverlevel.players().forEach((player) -> player.connection.send(serverlevel.clockManager().createFullSyncPacket(player))); // CraftBukkit - per-player
         }
+
+        // Neo: Allow mods to react to game rule state changes
+        net.neoforged.neoforge.event.EventHooks.onGameRuleChanged(serverlevel.getServer(), rule, value);
     }
 
     @Deprecated
@@ -2147,15 +_,15 @@
     }
 
     public SavedDataStorage getDataStorage() {
-        return this.savedDataStorage;
+        return this.overworld().getWorldContainerDataStorage(); // CraftBukkit
     }
 
     public TimerQueue<MinecraftServer> getScheduledEvents() {
-        return this.scheduledEvents;
+        return this.overworld().getScheduledEvents(); // CraftBukkit
     }
 
     public GameRules getGameRules() {
-        return this.gameRules;
+        return this.overworld().getGameRules(); // CraftBukkit
     }
 
     public boolean acceptsTransfers() {
