--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -5,17 +_,22 @@
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
+import com.mohistmc.youer.feature.YouerPlugin;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.jtracy.DiscontinuousFrame;
 import com.mojang.jtracy.TracyClient;
 import com.mojang.logging.LogUtils;
+import com.mojang.serialization.Dynamic;
+import com.mojang.serialization.Lifecycle;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import it.unimi.dsi.fastutil.objects.ObjectArraySet;
 import java.io.File;
 import java.io.IOException;
 import java.io.Writer;
+import java.lang.management.ManagementFactory;
 import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
 import java.net.Proxy;
 import java.nio.file.FileStore;
 import java.nio.file.Files;
@@ -30,12 +_,16 @@
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.Random;
 import java.util.Set;
 import java.util.UUID;
 import java.util.Map.Entry;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.locks.LockSupport;
 import java.util.function.BooleanSupplier;
@@ -43,6 +_,7 @@
 import java.util.function.Function;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
+import joptsimple.OptionSet;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.ReportType;
@@ -61,6 +_,8 @@
 import net.minecraft.core.registries.Registries;
 import net.minecraft.data.worldgen.features.MiscOverworldFeatures;
 import net.minecraft.gametest.framework.GameTestTicker;
+import net.minecraft.nbt.NbtException;
+import net.minecraft.nbt.ReportedNbtException;
 import net.minecraft.nbt.Tag;
 import net.minecraft.network.PacketProcessor;
 import net.minecraft.network.chat.ChatDecorator;
@@ -76,6 +_,8 @@
 import net.minecraft.resources.Identifier;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.bossevents.CustomBossEvents;
+import net.minecraft.server.dedicated.DedicatedServer;
+import net.minecraft.server.dedicated.DedicatedServerProperties;
 import net.minecraft.server.level.ChunkLoadCounter;
 import net.minecraft.server.level.ChunkMap;
 import net.minecraft.server.level.DemoMode;
@@ -115,6 +_,7 @@
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.TimeUtil;
 import net.minecraft.util.Util;
+import net.minecraft.util.datafix.DataFixers;
 import net.minecraft.util.debug.ServerDebugSubscribers;
 import net.minecraft.util.debugchart.SampleLogger;
 import net.minecraft.util.debugchart.TpsDebugDimensions;
@@ -133,6 +_,7 @@
 import net.minecraft.util.profiling.metrics.profiling.ServerMetricsSamplersProvider;
 import net.minecraft.util.profiling.metrics.storage.MetricsPersister;
 import net.minecraft.util.thread.ReentrantBlockableEventLoop;
+import net.minecraft.util.worldupdate.UpgradeProgress;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.RandomSequences;
 import net.minecraft.world.Stopwatches;
@@ -161,6 +_,7 @@
 import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.level.chunk.storage.ChunkIOErrorReporter;
 import net.minecraft.world.level.chunk.storage.RegionStorageInfo;
+import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.gamerules.GameRule;
 import net.minecraft.world.level.gamerules.GameRuleMap;
@@ -169,32 +_,48 @@
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.levelgen.PatrolSpawner;
 import net.minecraft.world.level.levelgen.PhantomSpawner;
+import net.minecraft.world.level.levelgen.WorldDimensions;
 import net.minecraft.world.level.levelgen.WorldGenSettings;
 import net.minecraft.world.level.levelgen.WorldOptions;
+import net.minecraft.world.level.levelgen.presets.WorldPresets;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
 import net.minecraft.world.level.saveddata.WeatherData;
 import net.minecraft.world.level.storage.CommandStorage;
 import net.minecraft.world.level.storage.DerivedLevelData;
 import net.minecraft.world.level.storage.LevelData;
+import net.minecraft.world.level.storage.LevelDataAndDimensions;
 import net.minecraft.world.level.storage.LevelResource;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import net.minecraft.world.level.storage.LevelSummary;
 import net.minecraft.world.level.storage.PlayerDataStorage;
+import net.minecraft.world.level.storage.PrimaryLevelData;
 import net.minecraft.world.level.storage.SavedDataStorage;
 import net.minecraft.world.level.storage.ServerLevelData;
 import net.minecraft.world.level.storage.WorldData;
 import net.minecraft.world.level.timers.TimerQueue;
+import net.minecraft.world.level.validation.ContentValidationException;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.scores.ScoreboardSaveData;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.CraftRegistry;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.util.CraftLocation;
+import org.bukkit.event.server.ServerLoadEvent;
+import org.bukkit.event.world.SpawnChangeEvent;
+import org.jline.terminal.Terminal;
+import org.jline.terminal.TerminalBuilder;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTask> implements CommandSource, ServerInfo, ChunkIOErrorReporter {
+public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTask> implements CommandSource, ServerInfo, ChunkIOErrorReporter, ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer { // Paper - rewrite chunk system
+    private static MinecraftServer SERVER; // Paper
     public static final Logger LOGGER = LogUtils.getLogger();
     public static final String VANILLA_BRAND = "vanilla";
     private static final float AVERAGE_TICK_TIME_SMOOTHING = 0.8F;
     private static final int TICK_STATS_SPAN = 100;
-    private static final long OVERLOADED_THRESHOLD_NANOS = 20L * TimeUtil.NANOSECONDS_PER_SECOND / 20L;
+    private static final long OVERLOADED_THRESHOLD_NANOS = 30L * TimeUtil.NANOSECONDS_PER_SECOND / 20L; // CraftBukkit
     private static final int OVERLOADED_TICKS_THRESHOLD = 20;
     private static final long OVERLOADED_WARNING_INTERVAL_NANOS = 10L * TimeUtil.NANOSECONDS_PER_SECOND;
     private static final int OVERLOADED_TICKS_WARNING_INTERVAL = 100;
@@ -218,7 +_,7 @@
     public final PlayerDataStorage playerDataStorage;
     private final SavedDataStorage savedDataStorage;
     private final List<Runnable> tickables = Lists.newArrayList();
-    private final GameRules gameRules;
+    //private final GameRules gameRules; // CraftBukkit - per world
     private MetricsRecorder metricsRecorder = InactiveMetricsRecorder.INSTANCE;
     private Consumer<ProfileResults> onMetricsRecordingStopped = results -> this.stopRecordingMetrics();
     private Consumer<Path> onMetricsRecordingFinished = ignored -> {};
@@ -237,13 +_,14 @@
     private Map<ResourceKey<Level>, ServerLevel> levels = Maps.newLinkedHashMap();
     private PlayerList playerList;
     private volatile boolean running = true;
+    private volatile boolean isRestarting = false; // Paper - flag to signify we're attempting to restart
     private boolean stopped;
     private int tickCount;
     private int ticksUntilAutosave = 6000;
     protected final Proxy proxy;
     private boolean onlineMode;
     private boolean preventProxyConnections;
-    private @Nullable String motd;
+    private net.kyori.adventure.text.Component motd; // Paper - Adventure
     private int playerIdleTimeout;
     private final long[] tickTimesNanos = new long[100];
     private long aggregatedTickTimesNanos = 0L;
@@ -271,7 +_,7 @@
     private @Nullable CommandStorage commandStorage;
     private final CustomBossEvents customBossEvents;
     private final RandomSequences randomSequences;
-    private final WeatherData weatherData;
+    // private final WeatherData weatherData; // CraftBukkit - per world
     private final ServerFunctionManager functionManager;
     private boolean enforceWhitelist;
     private boolean usingWhitelist;
@@ -291,12 +_,211 @@
     private final SuppressedExceptionCollector suppressedExceptions = new SuppressedExceptionCollector();
     private final DiscontinuousFrame tickFrame;
     private final PacketProcessor packetProcessor;
-    private final TimerQueue<MinecraftServer> scheduledEvents;
+    // private final TimerQueue<MinecraftServer> scheduledEvents; // CraftBukkit - per world
     private final ServerClockManager clockManager;
 
+    // CraftBukkit start
+    public final WorldLoader.DataLoadContext worldLoaderContext;
+    public org.bukkit.craftbukkit.CraftServer server;
+    public joptsimple.OptionSet options;
+    public org.bukkit.command.ConsoleCommandSender console;
+    public static int currentTick; // Paper - improve tick loop
+    public java.util.Queue<Runnable> processQueue = new java.util.concurrent.ConcurrentLinkedQueue<Runnable>();
+    public int autosavePeriod;
+    // Paper - don't store the vanilla dispatcher
+    public boolean forceTicks;
+    // CraftBukkit end
+    // Spigot start
+    public static final int TPS = 20;
+    public static final int TICK_TIME = 1000000000 / MinecraftServer.TPS;
+    // Spigot end
+    public volatile boolean hasFullyShutdown; // Paper - Improved watchdog support
+    public volatile boolean abnormalExit; // Paper - Improved watchdog support
+    public volatile Thread shutdownThread; // Paper - Improved watchdog support
+    public final io.papermc.paper.configuration.PaperConfigurations paperConfigurations; // Paper - add paper configuration files
+    public boolean isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
+    private final Set<String> pluginsBlockingSleep = new java.util.HashSet<>(); // Paper - API to allow/disallow tick sleeping
+    public static final long SERVER_INIT = System.nanoTime(); // Paper - Lag compensation
+    // Paper start - improve tick loop
+    public final ca.spottedleaf.moonrise.common.time.TickData tickTimes1s  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(1L));
+    public final ca.spottedleaf.moonrise.common.time.TickData tickTimes5s  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(5L));
+    public final ca.spottedleaf.moonrise.common.time.TickData tickTimes10s = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(10L));
+    public final ca.spottedleaf.moonrise.common.time.TickData tickTimes15s = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(15L));
+    public final ca.spottedleaf.moonrise.common.time.TickData tickTimes1m  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.MINUTES.toNanos(1L));
+    public final ca.spottedleaf.moonrise.common.time.TickData tickTimes5m  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.MINUTES.toNanos(5L));
+    public final ca.spottedleaf.moonrise.common.time.TickData tickTimes15m = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.MINUTES.toNanos(15L));
+
+    private final ca.spottedleaf.moonrise.common.time.Schedule tickSchedule = new ca.spottedleaf.moonrise.common.time.Schedule(0L);
+
+    private long lastTickStart;
+    private long currentTickStart;
+    private long scheduledTickStart;
+    private long taskExecutionTime;
+    /**
+     * The tickCount field is not incremented exactly where and when we want for our
+     * usage here.
+     * <p></p>
+     * There are two problems we need to fix:
+     * <ol>
+     *     <li>The tickCount field is not incremented when paused through integrated server.</li>
+     *     <li>The tickCount field is incremented after draining tasks (during server tick).</li>
+     * </ol>
+     * Our goal with the tick count here is to prevent executing tasks scheduled after the start
+     * of the current tick, which is marked by the task draining.
+     *
+     * @see #runAllTasksAtTickStart
+     */
+    private final java.util.concurrent.atomic.AtomicInteger tickTaskTickCount = new java.util.concurrent.atomic.AtomicInteger();
+    private final Object statsLock = new Object();
+    private @Nullable double[] tps;
+    private ca.spottedleaf.moonrise.common.time.TickData.@Nullable MSPTData msptData5s;
+
+    private void addTickTime(final ca.spottedleaf.moonrise.common.time.TickTime time) {
+        synchronized (this.statsLock) {
+            this.tickTimes1s.addDataFrom(time);
+            this.tickTimes5s.addDataFrom(time);
+            this.tickTimes10s.addDataFrom(time);
+            this.tickTimes15s.addDataFrom(time);
+            this.tickTimes1m.addDataFrom(time);
+            this.tickTimes5m.addDataFrom(time);
+            this.tickTimes15m.addDataFrom(time);
+            this.clearTickTimeStatistics();
+        }
+    }
+
+    private void clearTickTimeStatistics() {
+        this.msptData5s = null;
+        this.tps = null;
+    }
+
+    private static double getTPS(final ca.spottedleaf.moonrise.common.time.TickData tickData, final long tickInterval) {
+        final Double avg = tickData.getTPSAverage(null, tickInterval);
+        if (avg == null) {
+            return 1.0E9 / (double)tickInterval;
+        }
+
+        return avg;
+    }
+
+    public double[] getTPS() {
+        synchronized (this.statsLock) {
+            double[] tps = this.tps;
+            if (tps == null) {
+                tps = this.computeTPS();
+                this.tps = tps;
+            }
+            return tps.clone();
+        }
+    }
+
+    public ca.spottedleaf.moonrise.common.time.TickData.@Nullable MSPTData getMSPTData5s() {
+        synchronized (this.statsLock) {
+            if (this.msptData5s == null) {
+                this.msptData5s = this.tickTimes5s.getMSPTData(null, this.tickRateManager().nanosecondsPerTick());
+            }
+            return this.msptData5s;
+        }
+    }
+
+    public double[] computeTPS() {
+        final long interval = this.tickRateManager().nanosecondsPerTick();
+        return new double[] {
+                getTPS(this.tickTimes1m, interval),
+                getTPS(this.tickTimes5m, interval),
+                getTPS(this.tickTimes15m, interval)
+        };
+    }
+    // Paper end - improve tick loop
+    // Paper start - rewrite chunk system
+    private volatile Throwable chunkSystemCrash;
+
+    @Override
+    public final void moonrise$setChunkSystemCrash(final Throwable throwable) {
+        this.chunkSystemCrash = throwable;
+    }
+
+    private static final long CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME = 25L * 1000L; // 25us
+    private static final long MAX_CHUNK_EXEC_TIME = 1000L; // 1us
+    private static final long TASK_EXECUTION_FAILURE_BACKOFF = 5L * 1000L; // 5us
+
+    private long lastMidTickExecute;
+    private long lastMidTickExecuteFailure;
+
+    private boolean tickMidTickTasks() {
+        // give all worlds a fair chance at by targeting them all.
+        // if we execute too many tasks, that's fine - we have logic to correctly handle overuse of allocated time.
+        boolean executed = false;
+        for (final ServerLevel world : this.getAllLevels()) {
+            long currTime = System.nanoTime();
+            if (currTime - ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getLastMidTickFailure() <= TASK_EXECUTION_FAILURE_BACKOFF) {
+                continue;
+            }
+            if (!world.getChunkSource().pollTask()) {
+                // we need to back off if this fails
+                ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$setLastMidTickFailure(currTime);
+            } else {
+                executed = true;
+            }
+        }
+
+        return executed;
+    }
+
+    @Override
+    public final void moonrise$executeMidTickTasks() {
+        final long startTime = System.nanoTime();
+        if ((startTime - this.lastMidTickExecute) <= CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME || (startTime - this.lastMidTickExecuteFailure) <= TASK_EXECUTION_FAILURE_BACKOFF) {
+            // it's shown to be bad to constantly hit the queue (chunk loads slow to a crawl), even if no tasks are executed.
+            // so, backoff to prevent this
+            return;
+        }
+
+        for (;;) {
+            final boolean moreTasks = this.tickMidTickTasks();
+            final long currTime = System.nanoTime();
+            final long diff = currTime - startTime;
+
+            if (!moreTasks || diff >= MAX_CHUNK_EXEC_TIME) {
+                if (!moreTasks) {
+                    this.lastMidTickExecuteFailure = currTime;
+                }
+
+                // note: negative values reduce the time
+                long overuse = diff - MAX_CHUNK_EXEC_TIME;
+                if (overuse >= (10L * 1000L * 1000L)) { // 10ms
+                    // make sure something like a GC or dumb plugin doesn't screw us over...
+                    overuse = 10L * 1000L * 1000L; // 10ms
+                }
+
+                final double overuseCount = (double)overuse/(double)MAX_CHUNK_EXEC_TIME;
+                final long extraSleep = (long)Math.round(overuseCount*CHUNK_TASK_QUEUE_BACKOFF_MIN_TIME);
+
+                this.lastMidTickExecute = currTime + extraSleep;
+                return;
+            }
+        }
+    }
+
+    @Override
+    public final void moonrise$issueEmergencySave() {
+        LOGGER.warn("Performing emergency save...");
+        LOGGER.info("Saving all players...");
+        this.getPlayerList().saveAll();
+        LOGGER.info("Saved all players");
+        LOGGER.info("Saving all worlds...");
+        for (final ServerLevel world : this.getAllLevels()) {
+            LOGGER.info("Saving chunks in world '" + ca.spottedleaf.moonrise.common.util.WorldUtil.getWorldName(world) + "'...");
+            ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$issueEmergencySave();
+            LOGGER.info("Saved chunks in world '" + ca.spottedleaf.moonrise.common.util.WorldUtil.getWorldName(world) + "'...");
+        }
+        LOGGER.info("Saved all worlds");
+        LOGGER.warn("Performed emergency save");
+    }
+    // Paper end - rewrite chunk system
+
     public static <S extends MinecraftServer> S spin(Function<Thread, S> factory) {
         AtomicReference<S> serverReference = new AtomicReference<>();
-        Thread thread = new Thread(() -> serverReference.get().runServer(), "Server thread");
+        Thread thread = new Thread(net.neoforged.fml.util.thread.SidedThreadGroups.SERVER, () -> serverReference.get().runServer(), "Server thread");
         thread.setUncaughtExceptionHandler((t, e) -> LOGGER.error("Uncaught exception in server thread", e));
         if (Runtime.getRuntime().availableProcessors() > 4) {
             thread.setPriority(8);
@@ -309,6 +_,10 @@
     }
 
     public MinecraftServer(
+        // CraftBukkit start
+        joptsimple.OptionSet options,
+        WorldLoader.DataLoadContext worldLoaderContext,
+        // CraftBukkit end
         Thread serverThread,
         LevelStorageSource.LevelStorageAccess storageSource,
         PackRepository packRepository,
@@ -322,37 +_,36 @@
         NotificationManager notificationManager
     ) {
         super("Server", propagatesCrashes);
+        SERVER = this; // Paper - better singleton
         this.registries = worldStem.registries();
-        if (!this.registries.compositeAccess().lookupOrThrow(Registries.LEVEL_STEM).containsKey(LevelStem.OVERWORLD)) {
+        if (false && !this.registries.compositeAccess().lookupOrThrow(Registries.LEVEL_STEM).containsKey(LevelStem.OVERWORLD)) {
             throw new IllegalStateException("Missing Overworld dimension data");
         }
-
         this.savedDataStorage = new SavedDataStorage(storageSource.getLevelPath(LevelResource.DATA), fixerUpper, this.registries.compositeAccess());
         this.worldData = worldStem.worldDataAndGenSettings().data();
         this.worldGenSettings = worldStem.worldDataAndGenSettings().genSettings();
-        this.savedDataStorage.set(WorldGenSettings.TYPE, this.worldGenSettings);
+        //this.savedDataStorage.set(WorldGenSettings.TYPE, this.worldGenSettings);
         this.proxy = proxy;
         this.packRepository = packRepository;
         this.resources = new MinecraftServer.ReloadableResources(worldStem.resourceManager(), worldStem.dataPackResources());
         this.services = services;
-        this.connection = new ServerConnectionListener(this);
+        // this.connection = new ServerConnection(this); // Spigot
         this.tickRateManager = new ServerTickRateManager(this);
         this.levelLoadListener = levelLoadListener;
         this.storageSource = storageSource;
         this.playerDataStorage = storageSource.createPlayerStorage();
         this.randomSequences = this.savedDataStorage.computeIfAbsent(RandomSequences.TYPE);
-        this.weatherData = this.getDataStorage().computeIfAbsent(WeatherData.TYPE);
-        this.gameRules = new GameRules(this.worldData.enabledFeatures(), this.savedDataStorage.computeIfAbsent(GameRuleMap.TYPE));
-        gameRules.ifPresent(g -> this.gameRules.setAll(g, null));
+        //this.weatherData = this.getDataStorage().computeIfAbsent(WeatherData.TYPE);
+        //this.gameRules = new GameRules(this.worldData.enabledFeatures(), this.savedDataStorage.computeIfAbsent(GameRuleMap.TYPE));
         this.fixerUpper = fixerUpper;
         this.functionManager = new ServerFunctionManager(this, this.resources.managers.getFunctionLibrary());
         HolderGetter<Block> blockLookup = this.registries.compositeAccess().lookupOrThrow(Registries.BLOCK).filterFeatures(this.worldData.enabledFeatures());
         this.structureTemplateManager = new StructureTemplateManager(worldStem.resourceManager(), storageSource, fixerUpper, blockLookup);
         this.serverThread = serverThread;
         this.executor = Util.backgroundExecutor();
-        this.potionBrewing = PotionBrewing.bootstrap(this.worldData.enabledFeatures());
+        this.potionBrewing = PotionBrewing.bootstrap(this.worldData.enabledFeatures(), this.registryAccess());
         this.resources.managers.getRecipeManager().finalizeRecipeLoading(this.worldData.enabledFeatures());
-        this.fuelValues = FuelValues.vanillaBurnTimes(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+        this.fuelValues = net.neoforged.neoforge.common.DataMapHooks.populateFuelValues(this.registries.compositeAccess(), this.worldData.enabledFeatures());
         this.tickFrame = TracyClient.createDiscontinuousFrame("Server Tick");
         this.notificationManager = notificationManager;
         this.serverActivityMonitor = new ServerActivityMonitor(notificationManager, 30);
@@ -360,7 +_,14 @@
         this.clockManager = this.getDataStorage().computeIfAbsent(ServerClockManager.TYPE);
         this.clockManager.init(this);
         this.customBossEvents = this.savedDataStorage.computeIfAbsent(CustomBossEvents.TYPE);
-        this.scheduledEvents = this.savedDataStorage.computeIfAbsent(TimerQueue.TYPE);
+        // Paper - per-level scheduledEvents
+        // CraftBukkit start
+        this.options = options;
+        this.worldLoaderContext = worldLoaderContext;
+        io.papermc.paper.log.LogManagerShutdownThread.unhook(); // Paper - Improved watchdog support
+        Runtime.getRuntime().addShutdownHook(new org.bukkit.craftbukkit.util.ServerShutdownThread(this));
+        // CraftBukkit end
+        this.paperConfigurations = services.paper().configurations(); // Paper - add paper configuration files
     }
 
     protected abstract boolean initServer() throws IOException;
@@ -393,15 +_,15 @@
         };
     }
 
-    protected void loadLevel() {
+    protected void loadLevel(final String levelId) { // CraftBukkit
         boolean startedWorldLoadProfiling = !JvmProfiler.INSTANCE.isRunning()
             && SharedConstants.DEBUG_JFR_PROFILING_ENABLE_LEVEL_LOADING
             && JvmProfiler.INSTANCE.start(Environment.from(this));
         ProfiledDuration profiledDuration = JvmProfiler.INSTANCE.onWorldLoadedStarted();
-        this.worldData.setModdedInfo(this.getServerModName(), this.getModdedStatus().shouldReportAsModified());
-        this.createLevels();
-        this.forceDifficulty();
-        this.prepareLevels();
+        // Paper start - rework world loading process
+        io.papermc.paper.world.PaperWorldLoader loader = io.papermc.paper.world.PaperWorldLoader.create(this, levelId);
+        loader.loadInitialWorlds();
+        // Paper end - rework world loading process
         if (profiledDuration != null) {
             profiledDuration.finish(true);
         }
@@ -415,33 +_,151 @@
         }
     }
 
+    // Paper start - rework world loading process
+    protected void initPostWorld() {
+        // Paper start - Configurable player collision; Handle collideRule team for player collision toggle
+        final ServerScoreboard scoreboard = this.getScoreboard();
+        final java.util.Collection<String> toRemove = scoreboard.getPlayerTeams().stream().filter(team -> team.getName().startsWith("collideRule_")).map(net.minecraft.world.scores.PlayerTeam::getName).collect(java.util.stream.Collectors.toList());
+        for (String teamName : toRemove) {
+            scoreboard.removePlayerTeam(scoreboard.getPlayerTeam(teamName)); // Clean up after ourselves
+        }
+
+        if (!io.papermc.paper.configuration.GlobalConfiguration.get().collisions.enablePlayerCollisions) {
+            this.getPlayerList().collideRuleTeamName = org.apache.commons.lang3.StringUtils.left("collideRule_" + java.util.concurrent.ThreadLocalRandom.current().nextInt(), 16);
+            net.minecraft.world.scores.PlayerTeam collideTeam = scoreboard.addPlayerTeam(this.getPlayerList().collideRuleTeamName);
+            collideTeam.setSeeFriendlyInvisibles(false); // Because we want to mimic them not being on a team at all
+        }
+        // Paper end - Configurable player collision; Handle collideRule team for player collision toggle
+        this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
+        this.server.spark.registerCommandBeforePlugins(this.server); // Paper - spark
+        this.server.spark.enableAfterPlugins(this.server); // Paper - spark
+        io.papermc.paper.command.brigadier.PaperCommands.INSTANCE.setValid(); // Paper - reset invalid state for event fire below
+        io.papermc.paper.plugin.lifecycle.event.LifecycleEventRunner.INSTANCE.callReloadableRegistrarEvent(io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents.COMMANDS, io.papermc.paper.command.brigadier.PaperCommands.INSTANCE, org.bukkit.plugin.Plugin.class, io.papermc.paper.plugin.lifecycle.event.registrar.ReloadableRegistrarEvent.Cause.INITIAL); // Paper - call commands event for regular plugins
+        this.server.getCommandMap().registerServerAliases(); // Paper - relocate initial CommandMap#registerServerAliases() call
+        ((org.bukkit.craftbukkit.help.SimpleHelpMap) this.server.getHelpMap()).initializeCommands();
+        this.server.getPluginManager().callEvent(new org.bukkit.event.server.ServerLoadEvent(org.bukkit.event.server.ServerLoadEvent.LoadType.STARTUP));
+        this.connection.acceptConnections();
+    }
+    // Paper end - rework world loading process
+
     protected void forceDifficulty() {
     }
 
-    protected void createLevels() {
-        ServerLevelData levelData = this.worldData.overworldData();
-        boolean isDebug = this.worldData.isDebugWorld();
-        Registry<LevelStem> dimensions = this.registries.compositeAccess().lookupOrThrow(Registries.LEVEL_STEM);
-        WorldOptions worldOptions = this.worldGenSettings.options();
+    // Paper start - rework world loading process
+    public void createLevel(
+            LevelStem levelStem,
+            io.papermc.paper.world.PaperWorldLoader.WorldLoadingInfoAndData loading,
+            net.minecraft.world.level.storage.LevelDataAndDimensions.WorldDataAndGenSettings worldDataAndGenSettings
+    ) {
+        final WorldOptions worldOptions = worldDataAndGenSettings.genSettings().options();
+        final ResourceKey<Level> dimensionKey = loading.info().dimensionKey();
+        final SavedDataStorage savedDataStorage = new SavedDataStorage(this.storageSource.getDimensionPath(dimensionKey).resolve(LevelResource.DATA.id()), this.getFixerUpper(), this.registryAccess());
+        savedDataStorage.set(WorldGenSettings.TYPE, new WorldGenSettings(worldDataAndGenSettings.genSettings().options(), worldDataAndGenSettings.genSettings().dimensions()));
         long seed = worldOptions.seed();
         long biomeZoomSeed = BiomeManager.obfuscateSeed(seed);
         List<CustomSpawner> overworldCustomSpawners = ImmutableList.of(
-            new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(this.savedDataStorage)
-        );
-        LevelStem overworldData = dimensions.getValue(LevelStem.OVERWORLD);
-        ServerLevel overworld = new ServerLevel(
-            this, this.executor, this.storageSource, levelData, Level.OVERWORLD, overworldData, isDebug, biomeZoomSeed, overworldCustomSpawners, true
-        );
-        this.levels.put(Level.OVERWORLD, overworld);
-        this.scoreboard.load(this.savedDataStorage.computeIfAbsent(ScoreboardSaveData.TYPE).getData());
-        this.commandStorage = new CommandStorage(this.savedDataStorage);
-        this.stopwatches = this.savedDataStorage.computeIfAbsent(Stopwatches.TYPE);
+                new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(savedDataStorage) // Paper - save to world data
+        );
+        final org.bukkit.generator.ChunkGenerator chunkGenerator = this.server.getGenerator(loading.data().bukkitName());
+        org.bukkit.generator.BiomeProvider biomeProvider = this.server.getBiomeProvider(loading.data().bukkitName());
+        final org.bukkit.generator.WorldInfo worldInfo = new org.bukkit.craftbukkit.generator.CraftWorldInfo(
+                loading.data().bukkitName(),
+                org.bukkit.craftbukkit.util.CraftNamespacedKey.fromMinecraft(loading.info().dimensionKey().identifier()),
+                worldDataAndGenSettings.genSettings().options().seed(),
+                worldDataAndGenSettings.data().enabledFeatures(),
+                loading.info().environment(),
+                levelStem.type().value(),
+                levelStem.generator(),
+                this.registryAccess(),
+                loading.data().uuid()
+        );
+        if (biomeProvider == null && chunkGenerator != null) {
+            biomeProvider = chunkGenerator.getDefaultBiomeProvider(worldInfo);
+        }
+        ServerLevel serverLevel;
+        if (loading.info().stemKey() == LevelStem.OVERWORLD) {
+            serverLevel = new ServerLevel(
+                    this,
+                    this.executor,
+                    this.storageSource,
+                    worldDataAndGenSettings.genSettings(),
+                    dimensionKey,
+                    levelStem,
+                    worldDataAndGenSettings.data().isDebugWorld(),
+                    biomeZoomSeed,
+                    overworldCustomSpawners,
+                    true,
+                    loading.info().stemKey(),
+                    loading.info().environment(),
+                    chunkGenerator,
+                    biomeProvider,
+                    savedDataStorage,
+                    loading.data()
+            );
+            this.worldData = worldDataAndGenSettings.data();
+            this.worldData.setGameType(((net.minecraft.server.dedicated.DedicatedServer) this).getProperties().gameMode.get()); // From DedicatedServer.init
+            this.scoreboard.load(this.savedDataStorage.computeIfAbsent(ScoreboardSaveData.TYPE).getData());
+            this.commandStorage = new CommandStorage(this.savedDataStorage);
+            this.stopwatches = this.savedDataStorage.computeIfAbsent(Stopwatches.TYPE);
+            this.server.scoreboardManager = new org.bukkit.craftbukkit.scoreboard.CraftScoreboardManager(this, serverLevel.getScoreboard());
+        } else {
+            final List<CustomSpawner> spawners;
+            if (io.papermc.paper.configuration.GlobalConfiguration.get().misc.useDimensionTypeForCustomSpawners && levelStem.type().is(net.minecraft.world.level.dimension.BuiltinDimensionTypes.OVERWORLD)) {
+                spawners = overworldCustomSpawners;
+            } else {
+                spawners = Collections.emptyList();
+            }
+            serverLevel = new ServerLevel(
+                    this,
+                    this.executor,
+                    this.storageSource,
+                    worldDataAndGenSettings.genSettings(),
+                    dimensionKey,
+                    levelStem,
+                    this.worldData.isDebugWorld(),
+                    biomeZoomSeed,
+                    spawners,
+                    true,
+                    loading.info().stemKey(),
+                    loading.info().environment(),
+                    chunkGenerator,
+                    biomeProvider,
+                    savedDataStorage,
+                    loading.data()
+            );
+        }
+        this.addLevel(serverLevel);
+        this.initWorld(serverLevel, null);
+    }
+
+    public void initWorld(ServerLevel overworld, org.bukkit.@Nullable WorldCreator worldCreator) {
+        final net.minecraft.world.level.storage.ServerLevelData levelData = overworld.serverLevelData;
+        final WorldOptions worldOptions = overworld.worldGenSettings.options();
+        final boolean isDebug = this.worldData.isDebugWorld();
+        if (overworld.generator != null) {
+            overworld.getWorld().getPopulators().addAll(overworld.generator.getDefaultPopulators(overworld.getWorld()));
+        }
+        overworld.getWorldBorder().world = overworld;
+        overworld.getWorldBorder().setAbsoluteMaxSize(this.getAbsoluteMaxWorldSize());
+        this.getPlayerList().addWorldborderListener(overworld);
+        this.server.getPluginManager().callEvent(new org.bukkit.event.world.WorldInitEvent(overworld.getWorld()));
+        // Paper end - rework world loading process
         if (!levelData.isInitialized()) {
             try {
-                setInitialSpawn(overworld, levelData, worldOptions.generateBonusChest(), isDebug, this.levelLoadListener);
+                // Paper start - Allow direct setting spawn location
+                if (worldCreator != null && worldCreator.forcedSpawnPosition() != null) {
+                    levelData.setSpawn(LevelData.RespawnData.of(overworld.dimension(),
+                            io.papermc.paper.util.MCUtil.toBlockPos(worldCreator.forcedSpawnPosition()),
+                            Objects.requireNonNullElse(worldCreator.forcedSpawnYaw(), LevelData.RespawnData.DEFAULT.yaw()),
+                            Objects.requireNonNullElse(worldCreator.forcedSpawnPitch(), LevelData.RespawnData.DEFAULT.pitch())
+                    ));
+                } else {
+                    setInitialSpawn(overworld, levelData, worldOptions.generateBonusChest(), isDebug, overworld.levelLoadListener); // Paper - per world level load listener
+                }
+                // Paper end - Allow direct setting spawn location
                 levelData.setInitialized(true);
                 if (isDebug) {
-                    this.setupDebugLevel(this.worldData);
+                    this.setupDebugLevel(this.worldData, overworld); // Paper - per-level GameRules
                 }
             } catch (Throwable t) {
                 CrashReport report = CrashReport.forThrowable(t, "Exception initializing level");
@@ -458,25 +_,7 @@
         }
 
         GlobalPos focusPos = this.selectLevelLoadFocusPos();
-        this.levelLoadListener.updateFocus(focusPos.dimension(), ChunkPos.containing(focusPos.pos()));
-
-        for (Entry<ResourceKey<LevelStem>, LevelStem> entry : dimensions.entrySet()) {
-            ResourceKey<LevelStem> name = entry.getKey();
-            ServerLevel level;
-            if (name != LevelStem.OVERWORLD) {
-                ResourceKey<Level> dimension = ResourceKey.create(Registries.DIMENSION, name.identifier());
-                DerivedLevelData derivedLevelData = new DerivedLevelData(this.worldData, levelData);
-                level = new ServerLevel(
-                    this, this.executor, this.storageSource, derivedLevelData, dimension, entry.getValue(), isDebug, biomeZoomSeed, ImmutableList.of(), false
-                );
-                this.levels.put(dimension, level);
-            } else {
-                level = overworld;
-            }
-
-            level.getWorldBorder().setAbsoluteMaxSize(this.getAbsoluteMaxWorldSize());
-            this.getPlayerList().addWorldborderListener(level);
-        }
+        overworld.levelLoadListener.updateFocus(focusPos.dimension(), ChunkPos.containing(focusPos.pos())); // Paper - per world load listener
     }
 
     private static void setInitialSpawn(
@@ -488,7 +_,25 @@
             levelData.setSpawn(LevelData.RespawnData.of(level.dimension(), BlockPos.ZERO.above(80), 0.0F, 0.0F));
         } else {
             ServerChunkCache chunkSource = level.getChunkSource();
+            if (net.neoforged.neoforge.event.EventHooks.onCreateWorldSpawn(level, levelData)) return;
             ChunkPos spawnChunk = ChunkPos.containing(chunkSource.randomState().sampler().findSpawnPosition());
+
+            // CraftBukkit start
+            if (level.generator != null) {
+                Random rand = new Random(level.getSeed());
+                org.bukkit.Location spawn = level.generator.getFixedSpawnLocation(level.getWorld(), rand);
+
+                if (spawn != null) {
+                    if (spawn.getWorld() != level.getWorld()) {
+                        throw new IllegalStateException("Cannot set spawn point for " + levelData.getLevelName() + " to be in another world (" + spawn.getWorld().getName() + ")");
+                    } else {
+                        levelData.setSpawn(LevelData.RespawnData.of(level.dimension(), new BlockPos(spawn.getBlockX(), spawn.getBlockY(), spawn.getBlockZ()), spawn.getYaw(), spawn.getPitch()));
+                        return;
+                    }
+                }
+            }
+            // CraftBukkit end
+
             levelLoadListener.start(LevelLoadListener.Stage.PREPARE_GLOBAL_SPAWN, 0);
             levelLoadListener.updateFocus(level.dimension(), spawnChunk);
             int height = chunkSource.getGenerator().getSpawnHeight(level);
@@ -533,19 +_,22 @@
         }
     }
 
-    private void setupDebugLevel(WorldData worldData) {
+    private void setupDebugLevel(WorldData worldData, ServerLevel serverlevel) { // CraftBukkit - added level
         worldData.setDifficulty(Difficulty.PEACEFUL);
         worldData.setDifficultyLocked(true);
         ServerLevelData levelData = worldData.overworldData();
-        this.getGameRules().set(GameRules.ADVANCE_WEATHER, false, this);
-        this.clockManager.moveToTimeMarker(this.registryAccess().getOrThrow(WorldClocks.OVERWORLD), ClockTimeMarkers.NOON);
+        serverlevel.getGameRules().set(GameRules.ADVANCE_WEATHER, false, serverlevel); // CraftBukkit
+        serverlevel.clockManager().moveToTimeMarker(this.registryAccess().getOrThrow(WorldClocks.OVERWORLD), ClockTimeMarkers.NOON); // CraftBukkit
         levelData.setGameType(GameType.SPECTATOR);
     }
 
-    private void prepareLevels() {
+    // CraftBukkit start
+    public void prepareLevel(ServerLevel level) {
+        this.forceTicks = true;
+        // CraftBukkit end
         ChunkLoadCounter chunkLoadCounter = new ChunkLoadCounter();
 
-        for (ServerLevel level : this.levels.values()) {
+        if (true) { // CraftBukkit
             chunkLoadCounter.track(level, () -> {
                 TicketStorage savedTickets = level.getDataStorage().get(TicketStorage.TYPE);
                 if (savedTickets != null) {
@@ -554,17 +_,19 @@
             });
         }
 
-        this.levelLoadListener.start(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS, chunkLoadCounter.totalChunks());
+        level.levelLoadListener.start(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS, chunkLoadCounter.totalChunks()); // Paper - per world load listener
 
         do {
-            this.levelLoadListener.update(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS, chunkLoadCounter.readyChunks(), chunkLoadCounter.totalChunks());
-            this.nextTickTimeNanos = Util.getNanos() + PREPARE_LEVELS_DEFAULT_DELAY_NANOS;
-            this.waitUntilNextTick();
+            level.levelLoadListener.update(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS, chunkLoadCounter.readyChunks(), chunkLoadCounter.totalChunks()); // Paper - per world load listener
+            this.executeModerately(); // CraftBukkit
         } while (chunkLoadCounter.pendingChunks() > 0);
 
-        this.levelLoadListener.finish(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS);
-        this.updateMobSpawningFlags();
+        level.levelLoadListener.finish(LevelLoadListener.Stage.LOAD_INITIAL_CHUNKS); // Paper - per world load listener
+        level.setSpawnSettings(level.isSpawningMonsters()); // Paper - per level difficulty (from setDifficulty(ServerLevel, Difficulty, boolean))
         this.updateEffectiveRespawnData();
+        this.forceTicks = false; // CraftBukkit
+        //level.entityManager.tick(); // SPIGOT-6526: Load pending entities so they are available to the API // Paper - rewrite chunk system
+        new org.bukkit.event.world.WorldLoadEvent(level.getWorld()).callEvent(); // Paper - call WorldLoadEvent
     }
 
     protected GlobalPos selectLevelLoadFocusPos() {
@@ -586,7 +_,12 @@
     public abstract boolean shouldRconBroadcast();
 
     public boolean saveAllChunks(boolean silent, boolean flush, boolean force) {
-        this.scoreboard.storeToSaveDataIfDirty(this.getDataStorage().computeIfAbsent(ScoreboardSaveData.TYPE));
+        // Paper start - add close param
+        return this.saveAllChunks(silent, flush, force, false);
+    }
+    public boolean saveAllChunks(final boolean silent, final boolean flush, final boolean force, final boolean close) {
+        // Paper end - add close param
+        this.saveGlobalData(flush); // Paper - move to saveGlobalData()
         boolean result = false;
 
         for (ServerLevel level : this.getAllLevels()) {
@@ -598,6 +_,24 @@
             result = true;
         }
 
+        // Paper - move to saveGlobalData()
+
+        if (flush) {
+            for (ServerLevel level : this.getAllLevels()) {
+                String storageName = level.getChunkSource().chunkMap.getStorageName();
+                LOGGER.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", LEGACY_WORLD_NAMES_FOR_REALMS_LOG.getOrDefault(storageName, storageName));
+            }
+
+            LOGGER.info("ThreadedAnvilChunkStorage: All dimensions are saved");
+        }
+
+        return result;
+    }
+
+    // Paper start - split out global level data
+    public void saveGlobalData(final boolean flush) {
+        this.scoreboard.storeToSaveDataIfDirty(this.getDataStorage().computeIfAbsent(ScoreboardSaveData.TYPE));
+
         GameProfile singleplayerProfile = this.getSingleplayerProfile();
         this.storageSource.saveDataTag(this.worldData, singleplayerProfile == null ? null : singleplayerProfile.id());
         if (flush) {
@@ -605,18 +_,8 @@
         } else {
             this.savedDataStorage.scheduleSave();
         }
-
-        if (flush) {
-            for (ServerLevel level : this.getAllLevels()) {
-                String storageName = level.getChunkSource().chunkMap.getStorageName();
-                LOGGER.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", LEGACY_WORLD_NAMES_FOR_REALMS_LOG.getOrDefault(storageName, storageName));
-            }
-
-            LOGGER.info("ThreadedAnvilChunkStorage: All dimensions are saved");
-        }
-
-        return result;
     }
+    // Paper end - split out global level data
 
     public boolean saveEverything(boolean silent, boolean flush, boolean force) {
         try {
@@ -635,19 +_,49 @@
         this.stopServer();
     }
 
+    // CraftBukkit start
+    private boolean hasStopped = false;
+    private boolean hasLoggedStop = false; // Paper - Debugging
+    private final Object stopLock = new Object();
+    public final boolean hasStopped() {
+        synchronized (this.stopLock) {
+            return this.hasStopped;
+        }
+    }
+    // CraftBukkit end
+
     protected void stopServer() {
+        if (Thread.currentThread() == this.serverThread) this.executeAllRecentInternalTasks(); // Paper - execute tasks on stop
+        // CraftBukkit start - prevent double stopping on multiple threads
+        synchronized(stopLock) {
+            if (hasStopped) return;
+            hasStopped = true;
+        }
+        if (!this.hasLoggedStop && this.isDebugging()) io.papermc.paper.util.TraceUtil.dumpTraceForThread("Server stopped"); // Paper - Debugging
+        this.shutdownThread = Thread.currentThread(); // Paper - Improved watchdog support
+        org.spigotmc.WatchdogThread.doStop(); // Paper - Improved watchdog support
+        // CraftBukkit end
         this.packetProcessor.close();
         if (this.metricsRecorder.isRecording()) {
             this.cancelRecordingMetrics();
         }
 
         LOGGER.info("Stopping server");
+        Commands.COMMAND_SENDING_POOL.shutdownNow(); // Paper - Perf: Async command map building; Shutdown and don't bother finishing
+        // CraftBukkit start
+        if (this.server != null) {
+            this.server.spark.disable(); // Paper - spark
+            this.server.disablePlugins();
+            this.server.waitForAsyncTasksShutdown(); // Paper - Wait for Async Tasks during shutdown
+        }
+        // CraftBukkit end
         this.getConnection().stop();
         this.isSaving = true;
         if (this.playerList != null) {
             LOGGER.info("Saving players");
             this.playerList.saveAll();
-            this.playerList.removeAll();
+            this.playerList.removeAll(this.isRestarting); // Paper
+            try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
         }
 
         LOGGER.info("Saving worlds");
@@ -658,7 +_,7 @@
             }
         }
 
-        while (this.levels.values().stream().anyMatch(l -> l.getChunkSource().chunkMap.hasWork())) {
+        while (false && this.levels.values().stream().anyMatch(l -> l.getChunkSource().chunkMap.hasWork())) {
             this.nextTickTimeNanos = Util.getNanos() + TimeUtil.NANOSECONDS_PER_MILLISECOND;
 
             for (ServerLevel level : this.getAllLevels()) {
@@ -669,17 +_,14 @@
             this.waitUntilNextTick();
         }
 
-        this.saveAllChunks(false, true, false);
-
-        for (ServerLevel level : this.getAllLevels()) {
-            if (level != null) {
-                try {
-                    level.close();
-                } catch (IOException e) {
-                    LOGGER.error("Exception closing the level", e);
-                }
-            }
+        // Paper start - rewrite chunk system
+        // note: make sure we call deactivateTicketsOnClosing
+        for (final ServerLevel world : this.getAllLevels()) {
+            world.getChunkSource().deactivateTicketsOnClosing();
         }
+        // Paper end - rewrite chunk system
+
+        this.saveAllChunks(false, true, false, true); // Paper - rewrite chunk system
 
         this.isSaving = false;
         this.savedDataStorage.close();
@@ -690,6 +_,29 @@
         } catch (IOException e) {
             LOGGER.error("Failed to unlock level {}", this.storageSource.getLevelId(), e);
         }
+        // Spigot start
+        io.papermc.paper.util.MCUtil.ASYNC_EXECUTOR.shutdown(); // Paper
+        try {
+            io.papermc.paper.util.MCUtil.ASYNC_EXECUTOR.awaitTermination(30, java.util.concurrent.TimeUnit.SECONDS); // Paper
+        } catch (InterruptedException _) {} // Paper
+        if (org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly) {
+            LOGGER.info("Saving usercache.json");
+            this.services().nameToIdCache().save(false); // Paper - Perf: Async GameProfileCache saving
+        }
+        // Spigot end
+        // Paper start - rewrite chunk system
+        LOGGER.info("Waiting for all RegionFile I/O tasks to complete...");
+        ca.spottedleaf.moonrise.patches.chunk_system.io.MoonriseRegionFileIO.flush((MinecraftServer)(Object)this);
+        LOGGER.info("All RegionFile I/O tasks to complete");
+        if ((Object)this instanceof net.minecraft.server.dedicated.DedicatedServer) {
+            ca.spottedleaf.moonrise.common.util.MoonriseCommon.haltExecutors();
+        }
+        // Paper end - rewrite chunk system
+        // Paper start - Improved watchdog support - move final shutdown items here
+        Util.shutdownExecutors();
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStopped(this);
+        this.onServerExit();
+        // Paper end - Improved watchdog support - move final shutdown items here
     }
 
     public String getLocalIp() {
@@ -705,6 +_,14 @@
     }
 
     public void halt(boolean wait) {
+        // Paper start - allow passing of the intent to restart
+        this.safeShutdown(wait, false);
+    }
+    public void safeShutdown(boolean wait, final boolean isRestarting) {
+        this.isRestarting = isRestarting;
+        this.hasLoggedStop = true; // Paper - Debugging
+        if (this.isDebugging()) io.papermc.paper.util.TraceUtil.dumpTraceForThread("Server stopped"); // Paper - Debugging
+        // Paper end
         this.running = false;
         if (wait) {
             try {
@@ -715,37 +_,213 @@
         }
     }
 
+    // Paper start - improve tick loop
+    private void initTickSchedule() {
+        final long interval;
+        if (this.isPaused() || !this.tickRateManager.isSprinting()) {
+            interval = this.tickRateManager.nanosecondsPerTick();
+        } else {
+            interval = 0L;
+        }
+        this.tickSchedule.setNextPeriod(this.nextTickTimeNanos, interval);
+        this.lastTickStart = ca.spottedleaf.concurrentutil.util.TimeUtil.DEADLINE_NOT_SET;
+        this.scheduledTickStart = this.tickSchedule.getDeadline(interval);
+    }
+
+    private void recordEndOfTick() {
+        final long prevStart = this.lastTickStart;
+        final long currStart = this.currentTickStart;
+        this.lastTickStart = this.currentTickStart;
+        final long scheduledStart = this.scheduledTickStart;
+        this.scheduledTickStart = this.nextTickTimeNanos; // set scheduledStart for next tick
+
+        final long now = Util.getNanos();
+
+        final ca.spottedleaf.moonrise.common.time.TickTime time = new ca.spottedleaf.moonrise.common.time.TickTime(
+                prevStart,
+                scheduledStart,
+                currStart,
+                0L,
+                now,
+                0L,
+                this.taskExecutionTime,
+                0L,
+                false
+        );
+        this.taskExecutionTime = 0L;
+
+        this.addTickTime(time);
+    }
+
+    private void runAllTasksAtTickStart() {
+        this.startMeasuringTaskExecutionTime();
+
+        // note: To avoid possibly spinning forever, only execute tasks that are roughly available at the beginning
+        //       of this call. Packet processing and chunk system tasks are possibly always being queued.
+        final ProfilerFiller profiler = Profiler.get();
+        profiler.push("moonrise:run_all_tasks");
+
+        profiler.push("moonrise:run_all_server");
+        // avoid calling pollTask - we just want to execute queued tasks
+        final int currentTick = this.tickTaskTickCount.incrementAndGet();
+        final java.util.function.Predicate<net.minecraft.server.TickTask> taskPredicate = (final TickTask task) -> {
+            // only run tasks scheduled before the current tick - which we incremented above
+            return currentTick - task.getTick() > 0; // currentTick > tick accounting for overflow
+        };
+        while (this.runTaskIf(taskPredicate)) {
+            // execute small amounts of other tasks just in case the number of tasks we are
+            // draining is large - chunk system and packet processing may be latency sensitive
+
+            ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this).moonrise$executeMidTickTasks(); // Paper - rewrite chunk system
+            this.packetProcessor.executeSinglePacket();
+        }
+        profiler.popPush("moonrise:run_all_packets");
+        while (this.packetProcessor.executeSinglePacket()) {
+            // execute possibly latency sensitive chunk system tasks (see above)
+            ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this).moonrise$executeMidTickTasks(); // Paper - rewrite chunk system
+        }
+        profiler.popPush("moonrise:run_all_chunk");
+        // Paper start - rewrite chunk system
+        for (final ServerLevel world : this.getAllLevels()) {
+            profiler.push(world.toString() + " " + world.dimension().identifier()); // keep same formatting from regular tick, see tickChildren
+
+            // note: legacy tasks may expect a distance manager update
+            profiler.push("moonrise:distance_manager_update");
+            ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getChunkTaskScheduler().chunkHolderManager.processTicketUpdates();
+            profiler.popPush("moonrise:legacy_chunk_tasks");
+            world.getChunkSource().mainThreadProcessor.executeAllRecentInternalTasks();
+            profiler.popPush("moonrise:chunk_system_tasks");
+            ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getChunkTaskScheduler().executeAllRecentlyQueuedMainThreadTasks();
+            profiler.pop();
+
+            profiler.pop(); // world name
+        }
+        // Paper end - rewrite chunk system
+        profiler.pop(); // moonrise:run_all_chunk
+        profiler.pop(); // moonrise:run_all_tasks
+
+        this.finishMeasuringTaskExecutionTime();
+    }
+
+    private void recordTaskExecutionTimeWhileWaiting() {
+        final ProfilerFiller profiler = Profiler.get();
+
+        profiler.push("moonrise:execute_tasks_until_tick");
+        this.waitingForNextTick = true;
+        // implement waitForTasks
+        final boolean isLoggingEnabled = this.isTickTimeLoggingEnabled();
+        try {
+            final long deadline = this.nextTickTimeNanos;
+            for (;;) {
+                final long start = Util.getNanos();
+                if (start - deadline >= 0L) {
+                    // start is ahead of deadline
+                    break;
+                }
+
+                // execute tasks while there are tasks and there is time left
+                while (this.pollTask() && (Util.getNanos() - deadline < 0L));
+
+                final long now = Util.getNanos();
+
+                // record execution time
+                this.taskExecutionTime += (now - start);
+
+                // wait for unpark or deadline
+                final long toWait = deadline - now;
+                if (toWait > 0L) {
+                    LockSupport.parkNanos("waiting for tick or tasks", toWait);
+                    if (isLoggingEnabled) {
+                        this.idleTimeNanos += Util.getNanos() - now;
+                    }
+                } else {
+                    // done
+                    break;
+                }
+            }
+        } finally {
+            this.waitingForNextTick = false;
+        }
+        profiler.pop();
+    }
+    // Paper end - improve tick loop
+
     protected void runServer() {
         try {
             if (!this.initServer()) {
                 throw new IllegalStateException("Failed to initialize server");
             }
 
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarted(this);
             this.nextTickTimeNanos = Util.getNanos();
+            this.initTickSchedule(); // Paper - improve tick loop
             this.statusIcon = this.loadStatusIcon().orElse(null);
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
+
+            this.server.spark.enableBeforePlugins(); // Paper - spark
+            // Spigot start
+            // Paper start
+            LOGGER.info("Running delayed init tasks");
+            new io.papermc.paper.threadedregions.RegionizedServerInitEvent().callEvent(); // Call Folia init event
+            this.server.getScheduler().mainThreadHeartbeat(); // run all 1 tick delay tasks during init,
+            // this is going to be the first thing the tick process does anyway, so move done and run it after
+            // everything is init before watchdog tick.
+            // anything at 3+ won't be caught here but also will trip watchdog....
+            // tasks are default scheduled at -1 + delay, and first tick will tick at 1
+            final long actualDoneTimeMs = System.currentTimeMillis() - org.bukkit.craftbukkit.Main.BOOT_TIME.toEpochMilli(); // Paper - Improve startup message
+            LOGGER.info("Done ({})! For help, type \"help\"", String.format(java.util.Locale.ROOT, "%.3fs", actualDoneTimeMs / 1000.00D)); // Paper - Improve startup message
+            org.spigotmc.WatchdogThread.tick();
+            // Paper end
+            org.spigotmc.WatchdogThread.hasStarted = true; // Paper
+            // Paper start - Add onboarding message for initial server start
+            if (io.papermc.paper.configuration.GlobalConfiguration.isFirstStart) {
+                LOGGER.info("*************************************************************************************");
+                LOGGER.info("This is the first time you're starting this server.");
+                LOGGER.info("It's recommended you read our 'Getting Started' documentation for guidance.");
+                LOGGER.info("View this and more helpful information here: https://docs.papermc.io/paper/next-steps");
+                LOGGER.info("*************************************************************************************");
+            }
+            // Paper end - Add onboarding message for initial server start
+            // Paper start - Improve outdated version checking
+            if (System.getProperty("paper.disableStartupVersionCheck") == null && io.papermc.paper.configuration.GlobalConfiguration.get().updateChecker.enabled) {
+                CompletableFuture.runAsync(com.destroystokyo.paper.PaperVersionFetcher::getUpdateStatusStartupMessage, io.papermc.paper.util.MCUtil.ASYNC_EXECUTOR);
+            }
+            // Paper end - Improve outdated version checking
 
             while (this.running) {
-                long thisTickNanos;
+                final long tickStart = System.nanoTime(); // Paper - improve tick loop
+                long thisTickNanos; // Paper - improve tick loop - diff on change, expect this to be tick interval
                 if (!this.isPaused() && this.tickRateManager.isSprinting() && this.tickRateManager.checkShouldSprintThisTick()) {
                     thisTickNanos = 0L;
-                    this.nextTickTimeNanos = Util.getNanos();
-                    this.lastOverloadWarningNanos = this.nextTickTimeNanos;
+                    this.tickSchedule.setNextPeriod(tickStart, thisTickNanos); // Paper - improve tick loop
                 } else {
                     thisTickNanos = this.tickRateManager.nanosecondsPerTick();
-                    long behindTimeNanos = Util.getNanos() - this.nextTickTimeNanos;
-                    if (behindTimeNanos > OVERLOADED_THRESHOLD_NANOS + 20L * thisTickNanos
-                        && this.nextTickTimeNanos - this.lastOverloadWarningNanos >= OVERLOADED_WARNING_INTERVAL_NANOS + 100L * thisTickNanos) {
-                        long ticks = behindTimeNanos / thisTickNanos;
-                        LOGGER.warn(
-                            "Can't keep up! Is the server overloaded? Running {}ms or {} ticks behind",
-                            behindTimeNanos / TimeUtil.NANOSECONDS_PER_MILLISECOND,
-                            ticks
-                        );
-                        this.nextTickTimeNanos += ticks * thisTickNanos;
-                        this.lastOverloadWarningNanos = this.nextTickTimeNanos;
+                    // Paper start - improve tick loop
+                    // handle catchup logic
+                    final long ticksBehind = Math.max(1L, this.tickSchedule.getPeriodsAhead(thisTickNanos, tickStart));
+                    final long catchup = (long)Math.max(
+                            1,
+                            5 //ConfigHolder.getConfig().tickLoop.catchupTicks.getOrDefault(MoonriseConfig.TickLoop.DEFAULT_CATCHUP_TICKS).intValue()
+                    );
+
+                    // adjust ticksBehind so that it is not greater-than catchup
+                    if (ticksBehind - catchup > 0L) {
+                        final long difference = ticksBehind - catchup;
+                        this.tickSchedule.advanceBy(difference, thisTickNanos);
                     }
+
+                    // start next tick
+                    this.tickSchedule.advanceBy(1L, thisTickNanos);
+                    // Paper end - improve tick loop
                 }
+
+                this.nextTickTimeNanos = this.tickSchedule.getDeadline(thisTickNanos);
+                this.lastOverloadWarningNanos = this.nextTickTimeNanos;
+
+                this.currentTickStart = tickStart;
+                ++MinecraftServer.currentTick;
+                // Paper end - improve tick loop
 
                 boolean sprinting = thisTickNanos == 0L;
                 if (this.debugCommandProfilerDelayStart) {
@@ -753,7 +_,7 @@
                     this.debugCommandProfiler = new MinecraftServer.TimeProfiler(Util.getNanos(), this.tickCount);
                 }
 
-                this.nextTickTimeNanos += thisTickNanos;
+                // Paper - improve tick loop - done above
 
                 try (Profiler.Scope ignored = Profiler.use(this.createProfiler())) {
                     this.processPacketsAndTick(sprinting);
@@ -762,7 +_,7 @@
                     this.mayHaveDelayedTasks = true;
                     this.delayedTasksMaxNextTickTimeNanos = Math.max(Util.getNanos() + thisTickNanos, this.nextTickTimeNanos);
                     this.startMeasuringTaskExecutionTime();
-                    this.waitUntilNextTick();
+                    this.recordTaskExecutionTimeWhileWaiting(); // Paper - improve tick loop - record task execution here on MSPT
                     this.finishMeasuringTaskExecutionTime();
                     if (sprinting) {
                         this.tickRateManager.endTickWork();
@@ -777,6 +_,8 @@
                 this.isReady = true;
                 JvmProfiler.INSTANCE.onServerTick(this.smoothedTickTimeMillis);
             }
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStopping(this);
+            net.neoforged.neoforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
         } catch (Throwable t) {
             LOGGER.error("Encountered an unexpected exception", t);
             CrashReport report = constructOrExtractCrashReport(t);
@@ -788,6 +_,7 @@
                 LOGGER.error("We were unable to save this crash report to disk.");
             }
 
+            net.neoforged.neoforge.server.ServerLifecycleHooks.expectServerStopped(); // Forge: Has to come before MinecraftServer#onServerCrash to avoid race conditions
             this.onServerCrash(report);
         } finally {
             try {
@@ -795,8 +_,6 @@
                 this.stopServer();
             } catch (Throwable t) {
                 LOGGER.error("Exception stopping the server", t);
-            } finally {
-                this.onServerExit();
             }
         }
     }
@@ -848,7 +_,14 @@
     }
 
     private boolean haveTime() {
-        return this.runningTask() || Util.getNanos() < (this.mayHaveDelayedTasks ? this.delayedTasksMaxNextTickTimeNanos : this.nextTickTimeNanos);
+        // CraftBukkit start
+        return this.forceTicks || this.runningTask() || Util.getNanos() < (this.mayHaveDelayedTasks ? this.delayedTasksMaxNextTickTimeNanos : this.nextTickTimeNanos);
+    }
+
+    private void executeModerately() {
+        this.runAllTasks();
+        java.util.concurrent.locks.LockSupport.parkNanos("executing tasks", 1000L);
+        // CraftBukkit end
     }
 
     public NotificationManager notificationManager() {
@@ -956,6 +_,7 @@
     }
 
     protected void tickServer(BooleanSupplier haveTime) {
+        org.spigotmc.WatchdogThread.tick(); // Spigot
         long nano = Util.getNanos();
         int emptyTickThreshold = this.pauseWhenEmptySeconds() * 20;
         if (emptyTickThreshold > 0) {
@@ -971,21 +_,24 @@
                     this.autoSave();
                 }
 
+                this.server.getScheduler().mainThreadHeartbeat(); // CraftBukkit
                 this.tickConnection();
                 return;
             }
         }
 
         this.tickCount++;
+        net.neoforged.neoforge.event.EventHooks.fireServerTickPre(haveTime, this);
         this.tickRateManager.tick();
         this.tickChildren(haveTime);
         if (nano - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) {
             this.lastServerStatus = nano;
             this.status = this.buildServerStatus();
+            resetStatusCache(status);
         }
 
         this.ticksUntilAutosave--;
-        if (this.ticksUntilAutosave <= 0) {
+        if (this.autosavePeriod > 0 && this.ticksUntilAutosave <= 0) { // CraftBukkit
             this.autoSave();
         }
 
@@ -999,6 +_,7 @@
         this.smoothedTickTimeMillis = this.smoothedTickTimeMillis * 0.8F + (float)tickTime / (float)TimeUtil.NANOSECONDS_PER_MILLISECOND * 0.19999999F;
         this.logTickMethodTime(nano);
         profiler.pop();
+        net.neoforged.neoforge.event.EventHooks.fireServerTickPost(haveTime, this);
     }
 
     protected void processPacketsAndTick(boolean sprinting) {
@@ -1014,7 +_,7 @@
     }
 
     private void autoSave() {
-        this.ticksUntilAutosave = this.computeNextAutosaveInterval();
+        this.ticksUntilAutosave = this.autosavePeriod; // CraftBukkit
         LOGGER.debug("Autosave started");
         ProfilerFiller profiler = Profiler.get();
         profiler.push("save");
@@ -1029,6 +_,16 @@
         }
     }
 
+    private static final com.google.gson.Gson GSON = new com.google.gson.Gson();
+    private String cachedServerStatus; // Neo: cache the server status json in case a client spams requests
+    private void resetStatusCache(ServerStatus status) {
+        this.cachedServerStatus = GSON.toJson(ServerStatus.CODEC.encodeStart(com.mojang.serialization.JsonOps.INSTANCE, status)
+                .result().orElseThrow());
+    }
+    public String getStatusJson() {
+        return cachedServerStatus;
+    }
+
     private int computeNextAutosaveInterval() {
         float ticksPerSecond;
         if (this.tickRateManager.isSprinting()) {
@@ -1060,7 +_,8 @@
             Optional.of(players),
             Optional.of(ServerStatus.Version.current()),
             Optional.ofNullable(this.statusIcon),
-            this.enforceSecureProfile()
+            this.enforceSecureProfile(),
+            true //TODO Neo: Possible build a system which indicates what the status of the modded server is.
         );
     }
 
@@ -1087,12 +_,17 @@
     protected void tickChildren(BooleanSupplier haveTime) {
         ProfilerFiller profiler = Profiler.get();
         this.getPlayerList().getPlayers().forEach(playerx -> playerx.connection.suspendFlushing());
+        this.server.getScheduler().mainThreadHeartbeat(); // CraftBukkit
         profiler.push("commandFunctions");
         this.getFunctions().tick();
         profiler.pop();
         if (this.tickRateManager.runsNormally()) {
             profiler.push("clocks");
-            this.clockManager.tick();
+            // CraftBukkit start
+            for (ServerLevel serverlevel : this.getAllLevels()) {
+                serverlevel.clockManager().tick();
+            }
+            // CraftBukkit end
             profiler.pop();
         }
 
@@ -1105,9 +_,18 @@
         profiler.push("levels");
         this.updateEffectiveRespawnData();
 
-        for (ServerLevel level : this.getAllLevels()) {
+        // CraftBukkit start
+        // Run tasks that are waiting on processing
+        while (!processQueue.isEmpty()) {
+            processQueue.remove().run();
+        }
+        // CraftBukkit end
+
+        for (ServerLevel level : this.getWorldArray()) {
+            long tickStart = Util.getNanos();
             profiler.push(() -> level + " " + level.dimension().identifier());
             profiler.push("tick");
+            net.neoforged.neoforge.event.EventHooks.fireLevelTickPre(level, haveTime);
 
             try {
                 level.tick(haveTime);
@@ -1116,9 +_,11 @@
                 level.fillReportDetails(report);
                 throw new ReportedException(report);
             }
+            net.neoforged.neoforge.event.EventHooks.fireLevelTickPost(level, haveTime);
 
             profiler.pop();
             profiler.pop();
+            perWorldTickTimes.computeIfAbsent(level.dimension(), k -> new long[100])[this.tickCount % 100] = Util.getNanos() - tickStart;
         }
 
         profiler.popPush("connection");
@@ -1127,8 +_,8 @@
         this.playerList.tick();
         profiler.popPush("debugSubscribers");
         this.debugSubscribers.tick();
-        if (this.tickRateManager.runsNormally()) {
             profiler.popPush("gameTests");
+        if (net.neoforged.neoforge.gametest.GameTestHooks.isGametestEnabled() && this.tickRateManager.runsNormally()) {
             GameTestTicker.SINGLETON.tick();
         }
 
@@ -1149,7 +_,7 @@
         this.serverActivityMonitor.tick();
     }
 
-    protected void updateEffectiveRespawnData() {
+    public void updateEffectiveRespawnData() {
         LevelData.RespawnData respawnData = this.worldData.overworldData().getRespawnData();
         ServerLevel respawnLevel = this.findRespawnDimension();
         this.effectiveRespawnData = respawnLevel.getWorldBorderAdjustedRespawnData(respawnData);
@@ -1160,9 +_,16 @@
     }
 
     public void forceGameTimeSynchronization() {
+        if (this.overworld() == null) {
+            return; // Neo: Our ephemeral JUnit server has no overworld
+        }
         ProfilerFiller profiler = Profiler.get();
         profiler.push("timeSync");
-        this.playerList.broadcastAll(new ClientboundSetTimePacket(this.overworld().getGameTime(), Map.of()));
+        // CraftBukkit start
+        for (ServerLevel serverlevel : this.getAllLevels()) {
+            this.playerList.broadcastAll(new ClientboundSetTimePacket(serverlevel.getGameTime(), Map.of()), serverlevel);
+        }
+        // CraftBukkit end
         profiler.pop();
     }
 
@@ -1190,6 +_,22 @@
         return this.levels.get(dimension);
     }
 
+    // CraftBukkit start
+    public void addLevel(ServerLevel level) {
+        Map<ResourceKey<Level>, ServerLevel> oldLevels = this.levels;
+        Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
+        newLevels.put(level.dimension(), level);
+        this.levels = Collections.unmodifiableMap(newLevels);
+    }
+
+    public void removeLevel(ServerLevel level) {
+        Map<ResourceKey<Level>, ServerLevel> oldLevels = this.levels;
+        Map<ResourceKey<Level>, ServerLevel> newLevels = Maps.newLinkedHashMap(oldLevels);
+        newLevels.remove(level.dimension());
+        this.levels = Collections.unmodifiableMap(newLevels);
+    }
+    // CraftBukkit end
+
     public Set<ResourceKey<Level>> levelKeys() {
         return this.levels.keySet();
     }
@@ -1213,11 +_,11 @@
     }
 
     public String getServerModName() {
-        return "vanilla";
+        return net.neoforged.neoforge.internal.BrandingControl.getServerBranding();
     }
 
     public ServerClockManager clockManager() {
-        return this.clockManager;
+        return this.overworld().clockManager(); // CraftBukkit
     }
 
     public SystemReport fillSystemReport(SystemReport systemReport) {
@@ -1235,7 +_,7 @@
             () -> FeatureFlags.REGISTRY.toNames(this.worldData.enabledFeatures()).stream().map(Identifier::toString).collect(Collectors.joining(", "))
         );
         systemReport.setDetail("World Generation", () -> this.worldData.worldGenSettingsLifecycle().toString());
-        systemReport.setDetail("World Seed", () -> String.valueOf(this.worldGenSettings.options().seed()));
+        systemReport.setDetail("World Seed", (this.overworld() != null) ? String.valueOf(this.getWorldGenSettings().options().seed()) : "");
         systemReport.setDetail("Suppressed Exceptions", this.suppressedExceptions::dump);
         if (this.serverId != null) {
             systemReport.setDetail("Server Id", () -> this.serverId);
@@ -1369,10 +_,20 @@
 
     @Override
     public String getMotd() {
-        return this.motd;
+        return net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().serialize(this.motd); // Paper - Adventure
     }
 
     public void setMotd(String motd) {
+        // Paper start - Adventure
+        this.motd = net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection().deserializeOr(motd, net.kyori.adventure.text.Component.empty());
+    }
+
+    public net.kyori.adventure.text.Component motd() {
+        return this.motd;
+    }
+
+    public void motd(final net.kyori.adventure.text.Component motd) {
+        // Paper end - Adventure
         this.motd = motd;
     }
 
@@ -1411,7 +_,7 @@
     }
 
     public ServerConnectionListener getConnection() {
-        return this.connection;
+        return this.connection == null ? this.connection = new ServerConnectionListener(this) : this.connection; // Spigot
     }
 
     public boolean isReady() {
@@ -1515,9 +_,9 @@
 
     public CompletableFuture<Void> reloadResources(Collection<String> packsToEnable) {
         CompletableFuture<Void> result = CompletableFuture.<ImmutableList>supplyAsync(
-                () -> packsToEnable.stream()
-                    .map(this.packRepository::getPack)
-                    .filter(Objects::nonNull)
+                () -> this.packRepository
+                    .rebuildSelected(packsToEnable)
+                    .stream()
                     .map(Pack::open)
                     .collect(ImmutableList.toImmutableList()),
                 this
@@ -1547,6 +_,7 @@
             .thenAcceptAsync(newResources -> {
                 this.resources.close();
                 this.resources = newResources;
+                this.server.syncCommands(); // SPIGOT-5884: Lost on reload
                 this.packRepository.setSelected(packsToEnable);
                 WorldDataConfiguration newConfig = new WorldDataConfiguration(getSelectedPacks(this.packRepository, true), this.worldData.enabledFeatures());
                 this.worldData.setDataConfiguration(newConfig);
@@ -1556,7 +_,8 @@
                 this.getPlayerList().reloadResources();
                 this.functionManager.replaceLibrary(this.resources.managers.getFunctionLibrary());
                 this.structureTemplateManager.onResourceManagerReload(this.resources.resourceManager);
-                this.fuelValues = FuelValues.vanillaBurnTimes(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+                this.fuelValues = net.neoforged.neoforge.common.DataMapHooks.populateFuelValues(this.registries.compositeAccess(), this.worldData.enabledFeatures());
+                this.getPlayerList().getPlayers().forEach(this.getPlayerList()::sendPlayerPermissionLevel); //Forge: Fix newly added/modified commands not being sent to the client when commands reload.
             }, this);
         if (this.isSameThread()) {
             this.managedBlock(result::isDone);
@@ -1572,8 +_,10 @@
         FeatureFlagSet forcedFeatures = initMode ? FeatureFlagSet.of() : initialDataConfig.enabledFeatures();
         FeatureFlagSet allowedFeatures = initMode ? FeatureFlags.REGISTRY.allFlags() : initialDataConfig.enabledFeatures();
         packRepository.reload();
+        DataPackConfig.DEFAULT.addModPacks(net.neoforged.neoforge.common.CommonHooks.getModDataPacks());
+        dataPackConfig.addModPacks(net.neoforged.neoforge.common.CommonHooks.getModDataPacks());
         if (safeMode) {
-            return configureRepositoryWithSelection(packRepository, List.of("vanilla"), forcedFeatures, false);
+            return configureRepositoryWithSelection(packRepository, net.neoforged.neoforge.common.CommonHooks.getModDataPacksWithVanilla(), forcedFeatures, false);
         }
 
         Set<String> selected = Sets.newLinkedHashSet();
@@ -1620,6 +_,8 @@
             selected.add("vanilla");
         }
 
+        net.neoforged.neoforge.resource.ResourcePackLoader.reorderNewlyDiscoveredPacks(selected, dataPackConfig.getEnabled(), packRepository);
+
         return configureRepositoryWithSelection(packRepository, selected, forcedFeatures, true);
     }
 
@@ -1713,10 +_,20 @@
     }
 
     public void setRespawnData(LevelData.RespawnData respawnData) {
+        // CraftBukkit start
+        setRespawnData(respawnData, this.overworld());
+    }
+
+    public void setRespawnData(LevelData.RespawnData respawnData, ServerLevel world) {
+        // CraftBukkit end
         ServerLevelData levelData = this.worldData.overworldData();
         LevelData.RespawnData oldRespawnData = levelData.getRespawnData();
         if (!oldRespawnData.equals(respawnData)) {
             levelData.setSpawn(respawnData);
+            // CraftBukkit start - Notify anyone who's listening.
+            SpawnChangeEvent event = new SpawnChangeEvent(world.getWorld(), CraftLocation.toBukkit(oldRespawnData.pos(), world.getWorld(), oldRespawnData.yaw(), oldRespawnData.pitch()));
+            this.server.getPluginManager().callEvent(event);
+            // CraftBukkit end
             this.getPlayerList().broadcastAll(new ClientboundSetDefaultSpawnPositionPacket(respawnData));
             this.updateEffectiveRespawnData();
         }
@@ -1779,17 +_,19 @@
         return this.randomSequences;
     }
 
-    public void setWeatherParameters(int clearTime, int rainTime, boolean raining, boolean thundering) {
-        WeatherData weatherData = this.getWeatherData();
+    // Paper start - per-level WeatherData
+    public void setWeatherParameters(final ServerLevel level, final int clearTime, final int rainTime, final boolean raining, final boolean thundering) {
+        WeatherData weatherData = level.getWeatherData();
+        // Paper end - per-level WeatherData
         weatherData.setClearWeatherTime(clearTime);
         weatherData.setRainTime(rainTime);
         weatherData.setThunderTime(rainTime);
-        weatherData.setRaining(raining);
-        weatherData.setThundering(thundering);
+        weatherData.setRaining(raining, org.bukkit.event.weather.WeatherChangeEvent.Cause.COMMAND); // Paper - per-level WeatherData
+        weatherData.setThundering(thundering, org.bukkit.event.weather.ThunderChangeEvent.Cause.COMMAND); // Paper - per-level WeatherData
     }
 
     public WeatherData getWeatherData() {
-        return this.weatherData;
+        return this.overworld().getWeatherData(); // CraftBukkit
     }
 
     public boolean isEnforceWhitelist() {
@@ -1843,6 +_,31 @@
 
     public abstract boolean isSingleplayerOwner(NameAndId nameAndId);
 
+    private Map<ResourceKey<Level>, long[]> perWorldTickTimes = Maps.newIdentityHashMap();
+    @Nullable
+    public long[] getTickTime(ResourceKey<Level> dim) {
+        return perWorldTickTimes.get(dim);
+    }
+
+    @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+    public synchronized Map<ResourceKey<Level>, ServerLevel> forgeGetWorldMap() {
+        return this.levels;
+    }
+    private int worldArrayMarker = 0;
+    private int worldArrayLast = -1;
+    private ServerLevel[] worldArray;
+    @Deprecated //Forge Internal use Only, use to protect against concurrent modifications in the world tick loop.
+    public synchronized void markWorldsDirty() {
+        worldArrayMarker++;
+    }
+    private ServerLevel[] getWorldArray() {
+        if (worldArrayMarker == worldArrayLast && worldArray != null)
+            return worldArray;
+        worldArray = this.levels.values().stream().toArray(x -> new ServerLevel[x]);
+        worldArrayLast = worldArrayMarker;
+        return worldArray;
+    }
+
     public void dumpServerProperties(Path path) throws IOException {
     }
 
@@ -1937,6 +_,22 @@
         }
     }
 
+    // CraftBukkit start
+    public boolean isDebugging() {
+        return false;
+    }
+
+    @Deprecated
+    public static MinecraftServer getServer() {
+        return (Bukkit.getServer() instanceof CraftServer) ? ((CraftServer) Bukkit.getServer()).getServer() : null;
+    }
+
+    @Deprecated
+    public static RegistryAccess getDefaultRegistryAccess() {
+        return CraftRegistry.getMinecraftRegistry();
+    }
+    // CraftBukkit end
+
     private ProfilerFiller createProfiler() {
         if (this.willStartRecordingMetrics) {
             this.metricsRecorder = ActiveMetricsRecorder.createStarted(
@@ -2002,6 +_,10 @@
         return this.worldData;
     }
 
+    public MinecraftServer.ReloadableResources getServerResources() {
+         return resources;
+    }
+
     public RegistryAccess.Frozen registryAccess() {
         return this.registries.compositeAccess();
     }
@@ -2065,6 +_,13 @@
         }
     }
 
+    // CraftBukkit start
+    public final ExecutorService chatExecutor = Executors.newThreadPerTaskExecutor(Thread.ofVirtual()
+            .name("Async Chat Thread - #", 0)
+            .uncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(net.minecraft.server.MinecraftServer.LOGGER))
+            .factory());
+    // CraftBukkit end
+
     public ChatDecorator getChatDecorator() {
         return ChatDecorator.PLAIN;
     }
@@ -2073,6 +_,14 @@
         return true;
     }
 
+    public void handleCustomClickAction(Identifier id, Optional<Tag> payload, @Nullable ServerPlayer player, GameProfile profile) {
+        if (!net.neoforged.neoforge.common.CommonHooks.onCustomClickAction(player, profile, id, payload)) {
+            handleCustomClickAction(id, payload);
+        }
+    }
+
+    /// @deprecated Neo: Use the overload that has the player context and triggers the appropriate event.
+    @Deprecated
     public void handleCustomClickAction(Identifier id, Optional<Tag> payload) {
         LOGGER.debug("Received custom click action {} with payload {}", id, payload.orElse(null));
     }
@@ -2104,12 +_,12 @@
         return false;
     }
 
-    public <T> void onGameRuleChanged(GameRule<T> rule, T value) {
-        this.notificationManager().onGameRuleChanged(rule, value);
+    public static <T> void onGameRuleChanged(GameRule<T> rule, T value, ServerLevel serverlevel) { // CraftBukkit - per-world
+        serverlevel.getServer().notificationManager().onGameRuleChanged(rule, value);
         if (rule == GameRules.REDUCED_DEBUG_INFO) {
             byte event = (byte)((Boolean)value ? 22 : 23);
 
-            for (ServerPlayer player : this.getPlayerList().getPlayers()) {
+            for (ServerPlayer player : serverlevel.players()) {
                 player.connection.send(new ClientboundEntityEventPacket(player, event));
             }
         } else if (rule == GameRules.LIMITED_CRAFTING || rule == GameRules.IMMEDIATE_RESPAWN) {
@@ -2117,21 +_,23 @@
                 ? ClientboundGameEventPacket.LIMITED_CRAFTING
                 : ClientboundGameEventPacket.IMMEDIATE_RESPAWN;
             ClientboundGameEventPacket packet = new ClientboundGameEventPacket(eventType, (Boolean)value ? 1.0F : 0.0F);
-            this.getPlayerList().getPlayers().forEach(playerx -> playerx.connection.send(packet));
+            serverlevel.players().forEach(playerx -> playerx.connection.send(packet));
         } else if (rule == GameRules.LOCATOR_BAR) {
-            this.getAllLevels().forEach(level -> {
-                ServerWaypointManager waypointManager = level.getWaypointManager();
-                if ((Boolean)value) {
-                    level.players().forEach(waypointManager::updatePlayer);
-                } else {
-                    waypointManager.breakAllConnections();
-                }
-            });
+            ServerWaypointManager serverwaypointmanager = serverlevel.getWaypointManager();
+
+            if ((Boolean) value) {
+                serverlevel.players().forEach(serverwaypointmanager::updatePlayer);
+            } else {
+                serverwaypointmanager.breakAllConnections();
+            }
         } else if (rule == GameRules.SPAWN_MONSTERS) {
-            this.updateMobSpawningFlags();
+            serverlevel.setSpawnSettings(serverlevel.isSpawningMonsters()); // CraftBukkit - per-world
         } else if (rule == GameRules.ADVANCE_TIME) {
-            this.getPlayerList().broadcastAll(this.clockManager().createFullSyncPacket());
+            serverlevel.players().forEach((player) -> player.connection.send(serverlevel.clockManager().createFullSyncPacket(player))); // CraftBukkit - per-player
         }
+
+        // Neo: Allow mods to react to game rule state changes
+        net.neoforged.neoforge.event.EventHooks.onGameRuleChanged(serverlevel.getServer(), rule, value);
     }
 
     @Deprecated
@@ -2140,15 +_,17 @@
     }
 
     public SavedDataStorage getDataStorage() {
-        return this.savedDataStorage;
+        return this.overworld().getWorldContainerDataStorage(); // CraftBukkit
     }
 
+    @Deprecated(forRemoval = true) @io.papermc.paper.annotation.DoNotUse
     public TimerQueue<MinecraftServer> getScheduledEvents() {
-        return this.scheduledEvents;
+        throw new UnsupportedOperationException("Use ServerLevel.getScheduledEvents() instead"); // Paper
     }
 
+    @Deprecated(forRemoval = true) @io.papermc.paper.annotation.DoNotUse // Paper
     public GameRules getGameRules() {
-        return this.gameRules;
+        throw new UnsupportedOperationException("Use ServerLevel.getGameRules() instead"); // Paper
     }
 
     public boolean acceptsTransfers() {
@@ -2315,4 +_,53 @@
             };
         }
     }
+
+    // Paper start - Add tick times API and /mspt command
+    public static class TickTimes {
+        private final long[] times;
+
+        public TickTimes(int length) {
+            times = new long[length];
+        }
+
+        void add(int index, long time) {
+            times[index % times.length] = time;
+        }
+
+        public long[] getTimes() {
+            return times.clone();
+        }
+
+        public double getAverage() {
+            long total = 0L;
+            for (long value : times) {
+                total += value;
+            }
+            return ((double) total / (double) times.length) * 1.0E-6D;
+        }
+    }
+    // Paper end - Add tick times API and /mspt command
+
+    // Paper start - API to check if the server is sleeping
+    public boolean isTickPaused() {
+        return this.emptyTicks > 0 && this.emptyTicks >= this.pauseWhenEmptySeconds() * 20;
+    }
+
+    public void addPluginAllowingSleep(final String pluginName, final boolean value) {
+        if (!value) {
+            this.pluginsBlockingSleep.add(pluginName);
+        } else {
+            this.pluginsBlockingSleep.remove(pluginName);
+        }
+    }
+
+    private void removeDisabledPluginsBlockingSleep() {
+        if (this.pluginsBlockingSleep.isEmpty()) {
+            return;
+        }
+        this.pluginsBlockingSleep.removeIf(plugin -> (
+                !io.papermc.paper.plugin.manager.PaperPluginManagerImpl.getInstance().isPluginEnabled(plugin)
+        ));
+    }
+    // Paper end - API to check if the server is sleeping
 }
