--- a/net/minecraft/server/PlayerAdvancements.java
+++ b/net/minecraft/server/PlayerAdvancements.java
@@ -126,6 +_,7 @@
     }
 
     public void save() {
+        if (org.spigotmc.SpigotConfig.disableAdvancementSaving) return; // Spigot
         JsonElement json = this.codec.encodeStart(JsonOps.INSTANCE, this.asData()).getOrThrow();
 
         try {
@@ -143,6 +_,7 @@
         data.forEach((id, progress) -> {
             AdvancementHolder advancement = manager.get(id);
             if (advancement == null) {
+                if (!id.getNamespace().equals("minecraft")) return; // CraftBukkit
                 LOGGER.warn("Ignored advancement '{}' in progress file {} - it doesn't exist anymore?", id, this.playerSavePath);
             } else {
                 this.startProgress(advancement, progress);
@@ -163,6 +_,8 @@
     }
 
     public boolean award(AdvancementHolder holder, String criterion) {
+        // Forge: don't grant advancements for fake players
+        if (this.player instanceof net.neoforged.neoforge.common.util.FakePlayer) return false;
         boolean result = false;
         AdvancementProgress progress = this.getOrStartProgress(holder);
         boolean wasDone = progress.isDone();
@@ -170,12 +_,15 @@
             this.unregisterListeners(holder);
             this.progressChanged.add(holder);
             result = true;
+            net.neoforged.neoforge.event.EventHooks.onAdvancementProgressedEvent(this.player, holder, progress, criterion, net.neoforged.neoforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent.ProgressType.GRANT);
             if (!wasDone && progress.isDone()) {
+                this.player.level().getCraftServer().getPluginManager().callEvent(new org.bukkit.event.player.PlayerAdvancementDoneEvent(this.player.getBukkitEntity(), holder.toBukkit())); // CraftBukkit
                 holder.value().rewards().grant(this.player);
                 holder.value().display().ifPresent(display -> {
                     if (display.shouldAnnounceChat() && this.player.level().getGameRules().get(GameRules.SHOW_ADVANCEMENT_MESSAGES)) {
                         this.playerList.broadcastSystemMessage(display.getType().createAnnouncement(holder, this.player), false);
                     }
+                    net.neoforged.neoforge.event.EventHooks.onAdvancementEarnedEvent(this.player, holder);
                 });
             }
         }
@@ -195,6 +_,7 @@
             this.registerListeners(advancement);
             this.progressChanged.add(advancement);
             result = true;
+            net.neoforged.neoforge.event.EventHooks.onAdvancementProgressedEvent(this.player, advancement, progress, criterion, net.neoforged.neoforge.event.entity.player.AdvancementEvent.AdvancementProgressEvent.ProgressType.REVOKE);
         }
 
         if (wasDone && !progress.isDone()) {
