--- a/net/minecraft/server/RegistryLayer.java
+++ b/net/minecraft/server/RegistryLayer.java
@@ -12,7 +_,7 @@
     RELOADABLE;
 
     private static final List<RegistryLayer> VALUES = List.of(values());
-    private static final RegistryAccess.Frozen STATIC_ACCESS = RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY);
+    public static final RegistryAccess.Frozen STATIC_ACCESS = RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY);
 
     public static LayeredRegistryAccess<RegistryLayer> createRegistryAccess() {
         return new LayeredRegistryAccess<>(VALUES).replaceFrom(STATIC, STATIC_ACCESS);
