--- a/net/minecraft/server/ReloadableServerRegistries.java
+++ b/net/minecraft/server/ReloadableServerRegistries.java
@@ -61,14 +_,15 @@
         return CompletableFuture.supplyAsync(
             () -> {
                 WritableRegistry<T> writableregistry = new MappedRegistry<>(p_335741_.registryKey(), Lifecycle.experimental());
+                io.papermc.paper.registry.PaperRegistryAccess.instance().registerReloadableRegistry(p_335741_.registryKey(), writableregistry); // Paper - register reloadable registry
                 Map<ResourceLocation, JsonElement> map = new HashMap<>();
                 String s = Registries.elementsDirPath(p_335741_.registryKey());
                 SimpleJsonResourceReloadListener.scanDirectory(p_335893_, s, GSON, map);
                 map.forEach(
                     (p_335614_, p_335474_) -> p_335741_.deserialize(p_335614_, p_336173_, p_335474_)
                             .ifPresent(
-                                p_335683_ -> writableregistry.register(
-                                        ResourceKey.create(p_335741_.registryKey(), p_335614_), (T)p_335683_, DEFAULT_REGISTRATION_INFO
+                                p_335683_ ->io.papermc.paper.registry.PaperRegistryListenerManager.INSTANCE.registerWithListeners(writableregistry,
+                                        ResourceKey.create(p_335741_.registryKey(), p_335614_), (T)p_335683_, DEFAULT_REGISTRATION_INFO, new io.papermc.paper.registry.data.util.Conversions(p_336173_.lookupProvider)
                                     )
                             )
                 );
