--- a/net/minecraft/server/ServerAdvancementManager.java
+++ b/net/minecraft/server/ServerAdvancementManager.java
@@ -18,7 +_,7 @@
 
 public class ServerAdvancementManager {
     private static final Logger LOGGER = LogUtils.getLogger();
-    public Map<Identifier, AdvancementHolder> advancements;
+    private final Map<Identifier, AdvancementHolder> advancements;
     private final AdvancementTree tree;
 
     public ServerAdvancementManager(HolderLookup.Provider registries) {
