--- a/net/minecraft/server/ServerTickRateManager.java
+++ b/net/minecraft/server/ServerTickRateManager.java
@@ -66,6 +_,14 @@
         }
     }
 
+    public boolean stopSprinting(boolean sendLog) {
+        return stopSprinting();
+    }
+
+    private void finishTickSprint(boolean sendLog) {
+
+    }
+
     public boolean requestGameToSprint(int p_308951_) {
         boolean flag = this.remainingSprintTicks > 0L;
         this.sprintTimeSpend = 0L;
