--- a/net/minecraft/server/ServerAdvancementManager.java
+++ b/net/minecraft/server/ServerAdvancementManager.java
@@ -20,9 +_,13 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.util.HashMap;
+// CraftBukkit end
+
 public class ServerAdvancementManager extends SimpleJsonResourceReloadListener<Advancement> {
     private static final Logger LOGGER = LogUtils.getLogger();
-    public Map<Identifier, AdvancementHolder> advancements = Map.of();
+    public Map<Identifier, AdvancementHolder> advancements = new HashMap<>(); // CraftBukkit - SPIGOT-7734: mutable
     private AdvancementTree tree = new AdvancementTree();
     private final HolderLookup.Provider registries;
 
@@ -34,12 +_,18 @@
     protected void apply(Map<Identifier, Advancement> preparations, ResourceManager manager, ProfilerFiller profiler) {
         Builder<Identifier, AdvancementHolder> builder = ImmutableMap.builder();
         preparations.forEach((id, advancement) -> {
+            // Spigot start
+            if (org.spigotmc.SpigotConfig.disabledAdvancements != null && (org.spigotmc.SpigotConfig.disabledAdvancements.contains("*") || org.spigotmc.SpigotConfig.disabledAdvancements.contains(id.toString()) || org.spigotmc.SpigotConfig.disabledAdvancements.contains(id.getNamespace()))) {
+                return;
+            }
+            // Spigot end
             this.validate(id, advancement);
             builder.put(id, new AdvancementHolder(id, advancement));
         });
         this.advancements = builder.buildOrThrow();
         AdvancementTree tree = new AdvancementTree();
         tree.addAll(this.advancements.values());
+        LOGGER.info("Loaded {} advancements", tree.nodes().size()); // Paper - Improve logging and errors; moved from AdvancementTree#addAll
 
         for (AdvancementNode root : tree.roots()) {
             if (root.holder().value().display().isPresent()) {
