--- a/net/minecraft/server/ServerScoreboard.java
+++ b/net/minecraft/server/ServerScoreboard.java
@@ -51,17 +_,7 @@
     protected void onScoreChanged(ScoreHolder owner, Objective objective, Score score) {
         super.onScoreChanged(owner, objective, score);
         if (this.trackedObjectives.contains(objective)) {
-            this.server
-                .getPlayerList()
-                .broadcastAll(
-                    new ClientboundSetScorePacket(
-                        owner.getScoreboardName(),
-                        objective.getName(),
-                        score.value(),
-                        Optional.ofNullable(score.display()),
-                        Optional.ofNullable(score.numberFormat())
-                    )
-                );
+            this.broadcastAll(new ClientboundSetScorePacket(owner.getScoreboardName(), objective.getName(), score.value(), Optional.ofNullable(score.display()), Optional.ofNullable(score.numberFormat()))); // CraftBukkit
         }
 
         this.setDirty();
@@ -76,7 +_,7 @@
     @Override
     public void onPlayerRemoved(ScoreHolder player) {
         super.onPlayerRemoved(player);
-        this.server.getPlayerList().broadcastAll(new ClientboundResetScorePacket(player.getScoreboardName(), null));
+        this.broadcastAll(new ClientboundResetScorePacket(player.getScoreboardName(), (String) null)); // CraftBukkit
         this.setDirty();
     }
 
@@ -84,7 +_,7 @@
     public void onPlayerScoreRemoved(ScoreHolder player, Objective objective) {
         super.onPlayerScoreRemoved(player, objective);
         if (this.trackedObjectives.contains(objective)) {
-            this.server.getPlayerList().broadcastAll(new ClientboundResetScorePacket(player.getScoreboardName(), objective.getName()));
+            this.broadcastAll(new ClientboundResetScorePacket(player.getScoreboardName(), objective.getName())); // CraftBukkit
         }
 
         this.setDirty();
@@ -96,7 +_,7 @@
         super.setDisplayObjective(slot, objective);
         if (old != objective && old != null) {
             if (this.getObjectiveDisplaySlotCount(old) > 0) {
-                this.server.getPlayerList().broadcastAll(new ClientboundSetDisplayObjectivePacket(slot, objective));
+                this.broadcastAll(new ClientboundSetDisplayObjectivePacket(slot, objective)); // CraftBukkit
             } else {
                 this.stopTrackingObjective(old);
             }
@@ -104,7 +_,7 @@
 
         if (objective != null) {
             if (this.trackedObjectives.contains(objective)) {
-                this.server.getPlayerList().broadcastAll(new ClientboundSetDisplayObjectivePacket(slot, objective));
+                this.broadcastAll(new ClientboundSetDisplayObjectivePacket(slot, objective)); // CraftBukkit
             } else {
                 this.startTrackingObjective(objective);
             }
@@ -116,9 +_,7 @@
     @Override
     public boolean addPlayerToTeam(String player, PlayerTeam team) {
         if (super.addPlayerToTeam(player, team)) {
-            this.server
-                .getPlayerList()
-                .broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(team, player, ClientboundSetPlayerTeamPacket.Action.ADD));
+            this.broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(team, player, ClientboundSetPlayerTeamPacket.Action.ADD)); // CraftBukkit
             this.updatePlayerWaypoint(player);
             this.setDirty();
             return true;
@@ -130,7 +_,7 @@
     @Override
     public void removePlayerFromTeam(String player, PlayerTeam team) {
         super.removePlayerFromTeam(player, team);
-        this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(team, player, ClientboundSetPlayerTeamPacket.Action.REMOVE));
+        this.broadcastAll(ClientboundSetPlayerTeamPacket.createPlayerPacket(team, player, ClientboundSetPlayerTeamPacket.Action.REMOVE)); // CraftBukkit
         this.updatePlayerWaypoint(player);
         this.setDirty();
     }
@@ -145,7 +_,7 @@
     public void onObjectiveChanged(Objective objective) {
         super.onObjectiveChanged(objective);
         if (this.trackedObjectives.contains(objective)) {
-            this.server.getPlayerList().broadcastAll(new ClientboundSetObjectivePacket(objective, 2));
+            this.broadcastAll(new ClientboundSetObjectivePacket(objective, 2)); // CraftBukkit
         }
 
         this.setDirty();
@@ -164,14 +_,14 @@
     @Override
     public void onTeamAdded(PlayerTeam team) {
         super.onTeamAdded(team);
-        this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(team, true));
+        this.broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(team, true)); // CraftBukkit
         this.setDirty();
     }
 
     @Override
     public void onTeamChanged(PlayerTeam team) {
         super.onTeamChanged(team);
-        this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(team, false));
+        this.broadcastAll(ClientboundSetPlayerTeamPacket.createAddOrModifyPacket(team, false)); // CraftBukkit
         this.updateTeamWaypoints(team);
         this.setDirty();
     }
@@ -179,7 +_,7 @@
     @Override
     public void onTeamRemoved(PlayerTeam team) {
         super.onTeamRemoved(team);
-        this.server.getPlayerList().broadcastAll(ClientboundSetPlayerTeamPacket.createRemovePacket(team));
+        this.broadcastAll(ClientboundSetPlayerTeamPacket.createRemovePacket(team)); // CraftBukkit
         this.updateTeamWaypoints(team);
         this.setDirty();
     }
@@ -220,6 +_,7 @@
         List<Packet<?>> packets = this.getStartTrackingPackets(objective);
 
         for (ServerPlayer player : this.server.getPlayerList().getPlayers()) {
+            if (player.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
             for (Packet<?> packet : packets) {
                 player.connection.send(packet);
             }
@@ -245,6 +_,7 @@
         List<Packet<?>> packets = this.getStopTrackingPackets(objective);
 
         for (ServerPlayer player : this.server.getPlayerList().getPlayers()) {
+            if (player.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
             for (Packet<?> packet : packets) {
                 player.connection.send(packet);
             }
@@ -281,4 +_,14 @@
                 .forEach(player -> level.getWaypointManager().remakeConnections(player));
         }
     }
+
+    // CraftBukkit start - Send to players
+    private void broadcastAll(Packet packet) {
+        for (ServerPlayer serverplayer : (List<ServerPlayer>) this.server.getPlayerList().players) {
+            if (serverplayer.getBukkitEntity().getScoreboard().getHandle() == this) {
+                serverplayer.connection.send(packet);
+            }
+        }
+    }
+    // CraftBukkit end
 }
