--- a/net/minecraft/world/level/LevelAccessor.java
+++ b/net/minecraft/world/level/LevelAccessor.java
@@ -96,4 +96,6 @@
     default void gameEvent(ResourceKey<GameEvent> resourcekey, BlockPos blockpos, GameEvent.Context gameevent_context) {
         this.gameEvent(this.registryAccess().lookupOrThrow(Registries.GAME_EVENT).getOrThrow(resourcekey), blockpos, gameevent_context);
     }
+
+    net.minecraft.server.level.ServerLevel getMinecraftWorld(); // CraftBukkit
 }
