--- a/net/minecraft/world/level/TicketStorage.java
+++ b/net/minecraft/world/level/TicketStorage.java
@@ -30,6 +30,10 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.util.Objects;
+// CraftBukkit end
+
 public class TicketStorage extends SavedData {
 
     private static final int INITIAL_TICKET_LIST_CAPACITY = 4;
@@ -95,7 +99,7 @@
             Long2ObjectMap.Entry<List<Ticket>> long2objectmap_entry = (Entry) objectiterator.next();
             ChunkPos chunkpos = new ChunkPos(long2objectmap_entry.getLongKey());
 
-            for (Ticket ticket : (List) long2objectmap_entry.getValue()) {
+            for (Ticket ticket : long2objectmap_entry.getValue()) { // CraftBukkit - decompile error
                 biconsumer.accept(chunkpos, ticket);
             }
         }
@@ -108,7 +112,7 @@
         while (objectiterator.hasNext()) {
             Long2ObjectMap.Entry<List<Ticket>> long2objectmap_entry = (Entry) objectiterator.next();
 
-            for (Ticket ticket : (List) long2objectmap_entry.getValue()) {
+            for (Ticket ticket : long2objectmap_entry.getValue()) { // CraftBukkit - decompile error
                 this.addTicket(long2objectmap_entry.getLongKey(), ticket);
             }
         }
@@ -200,7 +204,7 @@
     }
 
     private static boolean isTicketSameTypeAndLevel(Ticket ticket, Ticket ticket1) {
-        return ticket1.getType() == ticket.getType() && ticket1.getTicketLevel() == ticket.getTicketLevel();
+        return ticket1.getType() == ticket.getType() && ticket1.getTicketLevel() == ticket.getTicketLevel() && Objects.equals(ticket.key, ticket1.key); // CraftBukkit
     }
 
     public int getTicketLevelAt(long i, boolean flag) {
@@ -401,7 +405,7 @@
         while (objectiterator.hasNext()) {
             Long2ObjectMap.Entry<List<Ticket>> long2objectmap_entry = (Entry) objectiterator.next();
 
-            for (Ticket ticket : (List) long2objectmap_entry.getValue()) {
+            for (Ticket ticket : long2objectmap_entry.getValue()) { // CraftBukkit - decompile error
                 if (ticket.getType() == tickettype) {
                     list.add(Pair.of(ticket, long2objectmap_entry.getLongKey()));
                 }
@@ -437,7 +441,7 @@
         while (objectiterator.hasNext()) {
             Long2ObjectMap.Entry<List<Ticket>> long2objectmap_entry = (Entry) objectiterator.next();
 
-            for (Ticket ticket : (List) long2objectmap_entry.getValue()) {
+            for (Ticket ticket : long2objectmap_entry.getValue()) { // CraftBukkit - decompile error
                 if (predicate.test(ticket)) {
                     longopenhashset.add(long2objectmap_entry.getLongKey());
                     break;
