--- a/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
@@ -9,6 +9,11 @@
 import net.minecraft.world.level.block.state.BlockState;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
+// CraftBukkit end
+
 public class DefaultRedstoneWireEvaluator extends RedstoneWireEvaluator {
 
     public DefaultRedstoneWireEvaluator(RedStoneWireBlock redstonewireblock) {
@@ -19,7 +24,16 @@
     public void updatePowerStrength(Level level, BlockPos blockpos, BlockState blockstate, @Nullable Orientation orientation, boolean flag) {
         int i = this.calculateTargetStrength(level, blockpos);
 
-        if ((Integer) blockstate.getValue(RedStoneWireBlock.POWER) != i) {
+        // CraftBukkit start
+        int oldPower = blockstate.getValue(RedStoneWireBlock.POWER);
+        if (oldPower != i) {
+            BlockRedstoneEvent event = new BlockRedstoneEvent(CraftBlock.at(level, blockpos), oldPower, i);
+            level.getCraftServer().getPluginManager().callEvent(event);
+
+            i = event.getNewCurrent();
+        }
+        if (oldPower != i) {
+            // CraftBukkit end
             if (level.getBlockState(blockpos) == blockstate) {
                 level.setBlock(blockpos, (BlockState) blockstate.setValue(RedStoneWireBlock.POWER, i), 2);
             }
