--- a/net/minecraft/client/renderer/texture/atlas/sources/PalettedPermutations.java
+++ b/net/minecraft/client/renderer/texture/atlas/sources/PalettedPermutations.java
@@ -93,7 +_,7 @@
         return MAP_CODEC;
     }
 
-    public record PalettedSpriteSupplier(LazyLoadedImage baseImage, Supplier<PaletteMapping> palette, Identifier permutationLocation)
+    private record PalettedSpriteSupplier(LazyLoadedImage baseImage, Supplier<PaletteMapping> palette, Identifier permutationLocation)
         implements SpriteSource.DiscardableLoader {
         @Override
         public @Nullable SpriteContents get(SpriteResourceLoader loader) {
