--- a/net/minecraft/client/renderer/texture/atlas/sources/Unstitcher.java
+++ b/net/minecraft/client/renderer/texture/atlas/sources/Unstitcher.java
@@ -67,13 +_,13 @@
         );
     }
 
-    public static class RegionInstance implements SpriteSource.DiscardableLoader {
+    private static class RegionInstance implements SpriteSource.DiscardableLoader {
         private final LazyLoadedImage image;
         private final Unstitcher.Region region;
         private final double xDivisor;
         private final double yDivisor;
 
-        public RegionInstance(LazyLoadedImage image, Unstitcher.Region region, double xDivisor, double yDivisor) {
+        private RegionInstance(LazyLoadedImage image, Unstitcher.Region region, double xDivisor, double yDivisor) {
             this.image = image;
             this.region = region;
             this.xDivisor = xDivisor;
