--- a/net/minecraft/world/entity/animal/camel/Camel.java
+++ b/net/minecraft/world/entity/animal/camel/Camel.java
@@ -286,6 +_,7 @@
         this.dashCooldown = 55;
         this.setDashing(true);
         this.hasImpulse = true;
+        net.neoforged.neoforge.common.CommonHooks.onLivingJump(this);
     }
 
     public boolean isDashing() {
@@ -450,10 +_,15 @@
         return SoundEvents.CAMEL_EAT;
     }
 
-    @Override
-    protected void actuallyHurt(DamageSource p_250410_, float p_251451_) {
+    // CraftBukkit start - void -> boolean
+    public boolean actuallyHurt(DamageSource damagesource, float f, org.bukkit.event.entity.EntityDamageEvent event) {
+        boolean damageResult = super.actuallyHurt(damagesource, f, event);
+        if (!damageResult) {
+            return false;
+        }
+        // CraftBukkit end
         this.standUpInstantly();
-        super.actuallyHurt(p_250410_, p_251451_);
+        return true; // CraftBukkit
     }
 
     @Override
