--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -42,6 +42,12 @@
 import net.minecraft.world.flag.FeatureFlags;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import net.minecraft.server.level.EntityPlayer;
+import org.bukkit.craftbukkit.CraftServerLinks;
+import org.bukkit.event.player.PlayerLinksSendEvent;
+// CraftBukkit end
+
 public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketListenerImpl implements ServerConfigurationPacketListener, TickablePacketListener {
 
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -57,8 +63,10 @@
     @Nullable
     private PrepareSpawnTask prepareSpawnTask;
 
-    public ServerConfigurationPacketListenerImpl(MinecraftServer minecraftserver, NetworkManager networkmanager, CommonListenerCookie commonlistenercookie) {
-        super(minecraftserver, networkmanager, commonlistenercookie);
+    // CraftBukkit start
+    public ServerConfigurationPacketListenerImpl(MinecraftServer minecraftserver, NetworkManager networkmanager, CommonListenerCookie commonlistenercookie, EntityPlayer player) {
+        super(minecraftserver, networkmanager, commonlistenercookie, player);
+        // CraftBukkit end
         this.gameProfile = commonlistenercookie.gameProfile();
         this.clientInformation = commonlistenercookie.clientInformation();
     }
@@ -87,6 +95,12 @@
     public void startConfiguration() {
         this.send(new ClientboundCustomPayloadPacket(new BrandPayload(this.server.getServerModName())));
         ServerLinks serverlinks = this.server.serverLinks();
+        // CraftBukkit start
+        CraftServerLinks wrapper = new CraftServerLinks(serverlinks);
+        PlayerLinksSendEvent event = new PlayerLinksSendEvent(player.getBukkitEntity(), wrapper);
+        player.getBukkitEntity().getServer().getPluginManager().callEvent(event);
+        serverlinks = wrapper.getServerLinks();
+        // CraftBukkit end
 
         if (!serverlinks.isEmpty()) {
             this.send(new ClientboundServerLinksPacket(serverlinks.untrust()));
@@ -105,7 +119,7 @@
     }
 
     public void returnToWorld() {
-        this.prepareSpawnTask = new PrepareSpawnTask(this.server, new NameAndId(this.gameProfile));
+        this.prepareSpawnTask = new PrepareSpawnTask(this.server, this.player); // CraftBukkit
         this.configurationTasks.add(this.prepareSpawnTask);
         this.configurationTasks.add(new JoinWorldTask());
         this.startNextTask();
@@ -179,13 +193,17 @@
                 return;
             }
 
-            IChatBaseComponent ichatbasecomponent = playerlist.canPlayerLogin(this.connection.getRemoteAddress(), new NameAndId(this.gameProfile));
+            IChatBaseComponent ichatbasecomponent = null; // CraftBukkit - login checks already completed
 
             if (ichatbasecomponent != null) {
                 this.disconnect(ichatbasecomponent);
                 return;
             }
 
+            // CraftBukkit start
+            EntityPlayer entityplayer = this.player;
+            entityplayer.updateOptions(this.clientInformation);
+            // CraftBukkit end
             ((PrepareSpawnTask) Objects.requireNonNull(this.prepareSpawnTask)).spawnPlayer(this.connection, this.createCookie(this.clientInformation));
         } catch (Exception exception) {
             ServerConfigurationPacketListenerImpl.LOGGER.error("Couldn't place player in world", exception);
