--- a/net/minecraft/server/network/ServerConnection.java
+++ b/net/minecraft/server/network/ServerConnection.java
@@ -104,9 +104,19 @@
                     networkmanager.configurePacketHandler(channelpipeline);
                     networkmanager.setListenerForServerboundHandshake(new HandshakeListener(ServerConnection.this.server, networkmanager));
                 }
-            }).group(eventloopgroup).localAddress(inetaddress, i)).bind().syncUninterruptibly());
+            }).group(eventloopgroup).localAddress(inetaddress, i)).option(ChannelOption.AUTO_READ, false).bind().syncUninterruptibly()); // CraftBukkit
+        }
+    }
+
+    // CraftBukkit start
+    public void acceptConnections() {
+        synchronized (this.channels) {
+            for (ChannelFuture future : this.channels) {
+                future.channel().config().setAutoRead(true);
+            }
         }
     }
+    // CraftBukkit end
 
     public SocketAddress startMemoryChannel() {
         ChannelFuture channelfuture;
