--- a/net/minecraft/world/item/consume_effects/ApplyStatusEffectsConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/ApplyStatusEffectsConsumeEffect.java
@@ -12,6 +12,10 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.World;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+// CraftBukkit end
+
 public record ApplyStatusEffectsConsumeEffect(List<MobEffect> effects, float probability) implements ConsumeEffect {
 
     public static final MapCodec<ApplyStatusEffectsConsumeEffect> CODEC = RecordCodecBuilder.mapCodec((instance) -> {
@@ -37,14 +41,14 @@
     }
 
     @Override
-    public boolean apply(World world, ItemStack itemstack, EntityLiving entityliving) {
+    public boolean apply(World world, ItemStack itemstack, EntityLiving entityliving, EntityPotionEffectEvent.Cause cause) { // CraftBukkit
         if (entityliving.getRandom().nextFloat() >= this.probability) {
             return false;
         } else {
             boolean flag = false;
 
             for (MobEffect mobeffect : this.effects) {
-                if (entityliving.addEffect(new MobEffect(mobeffect))) {
+                if (entityliving.addEffect(new MobEffect(mobeffect), cause)) { // CraftBukkit
                     flag = true;
                 }
             }
