--- a/com/mojang/blaze3d/vertex/PoseStack.java
+++ b/com/mojang/blaze3d/vertex/PoseStack.java
@@ -1,6 +_,5 @@
 package com.mojang.blaze3d.vertex;
 
-import com.mojang.math.Axis;
 import com.mojang.math.MatrixUtil;
 import com.mojang.math.Transformation;
 import java.util.ArrayList;
@@ -41,18 +_,10 @@
         this.last().scale(xScale, yScale, zScale);
     }
 
-    public void rotate(Quaternionfc by) {
+    public void mulPose(Quaternionfc by) {
         this.last().rotate(by);
     }
 
-    public void rotate(Axis axis, float angle) {
-        this.last().rotate(axis, angle);
-    }
-
-    public void rotateDegrees(Axis axis, float angle) {
-        this.last().rotateDegrees(axis, angle);
-    }
-
     public void rotateAround(Quaternionfc rotation, float pivotX, float pivotY, float pivotZ) {
         this.last().rotateAround(rotation, pivotX, pivotY, pivotZ);
     }
@@ -147,16 +_,6 @@
         public void rotate(Quaternionfc by) {
             this.pose.rotate(by);
             this.normal.rotate(by);
-        }
-
-        public void rotate(Axis axis, float angle) {
-            axis.rotate(this.pose, angle);
-            axis.rotate(this.normal, angle);
-        }
-
-        public void rotateDegrees(Axis axis, float angle) {
-            axis.rotateDegrees(this.pose, angle);
-            axis.rotateDegrees(this.normal, angle);
         }
 
         public void rotateAround(Quaternionfc rotation, float pivotX, float pivotY, float pivotZ) {
