--- a/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
+++ b/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
@@ -60,6 +_,7 @@
                                     return false;
                                 } else {
                                     mutablelong.setValue(p_258294_ + 20L + (long)p_258292_.getRandom().nextInt(20));
+                                    if (p_258293_.getNavigation().isStuck()) mutablelong.add(200); // Paper - Perf: Wait an additional 10s to check again if they're stuck
                                     PoiManager poimanager = p_258292_.getPoiManager();
                                     long2objectmap.long2ObjectEntrySet().removeIf(p_22338_ -> !p_22338_.getValue().isStillValid(p_258294_));
                                     Predicate<BlockPos> predicate = p_258266_ -> {
