--- a/net/minecraft/world/entity/ai/behavior/AssignProfessionFromJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/AssignProfessionFromJobSite.java
@@ -10,6 +_,9 @@
 import net.minecraft.world.entity.ai.village.poi.PoiType;
 import net.minecraft.world.entity.npc.villager.Villager;
 import net.minecraft.world.entity.npc.villager.VillagerProfession;
+import org.bukkit.craftbukkit.entity.CraftVillager;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerCareerChangeEvent;
 
 public class AssignProfessionFromJobSite {
     public static BehaviorControl<Villager> create() {
@@ -19,7 +_,7 @@
                     i,
                     (potentialJobSite, jobSite) -> (level, body, timestamp) -> {
                         GlobalPos pos = i.get(potentialJobSite);
-                        if (!pos.pos().closerToCenterThan(body.position(), 2.0)) {
+                        if (!pos.pos().closerToCenterThan(body.position(), 2.0) && !body.assignProfessionWhenSpawned()) {
                             return false;
                         }
 
@@ -40,7 +_,14 @@
                                     .findFirst()
                             )
                             .ifPresent(profession -> {
-                                body.setVillagerData(body.getVillagerData().withProfession(profession));
+                                        // CraftBukkit start - Fire VillagerCareerChangeEvent where Villager gets employed
+                                        VillagerCareerChangeEvent event = CraftEventFactory.callVillagerCareerChangeEvent(body, CraftVillager.CraftProfession.minecraftHolderToBukkit(profession), VillagerCareerChangeEvent.ChangeReason.EMPLOYED);
+                                        if (event.isCancelled()) {
+                                            return;
+                                        }
+
+                                        body.setVillagerData(body.getVillagerData().withProfession(CraftVillager.CraftProfession.bukkitToMinecraftHolder(event.getProfession())));
+                                        // CraftBukkit end
                                 body.refreshBrain(level);
                             });
                         return true;
