--- a/net/minecraft/world/entity/ai/behavior/FollowTemptation.java
+++ b/net/minecraft/world/entity/ai/behavior/FollowTemptation.java
@@ -37,6 +_,7 @@
             builder.put(MemoryModuleType.LOOK_TARGET, MemoryStatus.REGISTERED);
             builder.put(MemoryModuleType.WALK_TARGET, MemoryStatus.REGISTERED);
             builder.put(MemoryModuleType.TEMPTATION_COOLDOWN_TICKS, MemoryStatus.VALUE_ABSENT);
+            builder.put(MemoryModuleType.IS_TEMPTED, MemoryStatus.VALUE_ABSENT);
             builder.put(MemoryModuleType.TEMPTING_PLAYER, MemoryStatus.VALUE_PRESENT);
             builder.put(MemoryModuleType.BREED_TARGET, MemoryStatus.VALUE_ABSENT);
             builder.put(MemoryModuleType.IS_PANICKING, MemoryStatus.VALUE_ABSENT);
@@ -66,9 +_,14 @@
             && !body.getBrain().hasMemoryValue(MemoryModuleType.IS_PANICKING);
     }
 
+    protected void start(ServerLevel level, PathfinderMob body, long timestamp) {
+        body.getBrain().setMemory(MemoryModuleType.IS_TEMPTED, true);
+    }
+
     protected void stop(ServerLevel level, PathfinderMob body, long timestamp) {
         Brain<?> brain = body.getBrain();
         brain.setMemory(MemoryModuleType.TEMPTATION_COOLDOWN_TICKS, 100);
+        brain.eraseMemory(MemoryModuleType.IS_TEMPTED);
         brain.eraseMemory(MemoryModuleType.WALK_TARGET);
         brain.eraseMemory(MemoryModuleType.LOOK_TARGET);
     }
