--- a/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
+++ b/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
@@ -53,9 +_,16 @@
                             Node toNode = path.getNextNode();
                             BlockPos fromPos = fromNode.asBlockPos();
                             BlockState fromState = level.getBlockState(fromPos);
-                            if (fromState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)) {
+                            if (fromState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)&& !DoorBlock.requiresRedstone(body.level(), fromState, fromPos)) { // Purpur - Option to make doors require redstone
                                 DoorBlock fromBlock = (DoorBlock)fromState.getBlock();
                                 if (!fromBlock.isOpen(fromState)) {
+                                    // CraftBukkit start - entities opening doors
+                                    org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(body.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(body.level(), fromPos));
+                                    body.level().getCraftServer().getPluginManager().callEvent(event);
+                                    if (event.isCancelled()) {
+                                        return false;
+                                    }
+                                    // CraftBukkit end
                                     fromBlock.setOpen(body, level, fromState, fromPos, true);
                                 }
 
@@ -64,9 +_,16 @@
 
                             BlockPos toPos = toNode.asBlockPos();
                             BlockState toState = level.getBlockState(toPos);
-                            if (toState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)) {
+                            if (toState.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock) && !DoorBlock.requiresRedstone(body.level(), toState, toPos)) { // Purpur - Option to make doors require redstone
                                 DoorBlock door = (DoorBlock)toState.getBlock();
                                 if (!door.isOpen(toState)) {
+                                    // CraftBukkit start - entities opening doors
+                                    org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(body.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(body.level(), toPos));
+                                    body.level().getCraftServer().getPluginManager().callEvent(event);
+                                    if (event.isCancelled()) {
+                                        return false;
+                                    }
+                                    // CraftBukkit end
                                     door.setOpen(body, level, toState, toPos, true);
                                     doors = rememberDoorToClose(doorsMemory, doors, level, toPos);
                                 }
@@ -105,7 +_,7 @@
                     iterator.remove();
                 } else {
                     BlockState state = level.getBlockState(doorPos);
-                    if (!state.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock)) {
+                    if (!state.is(BlockTags.MOB_INTERACTABLE_DOORS, s -> s.getBlock() instanceof DoorBlock) || DoorBlock.requiresRedstone(body.level(), state, doorPos)) { // Purpur - Option to make doors require redstone
                         iterator.remove();
                     } else {
                         DoorBlock block = (DoorBlock)state.getBlock();
