--- a/net/minecraft/world/entity/ai/behavior/MeleeAttack.java
+++ b/net/minecraft/world/entity/ai/behavior/MeleeAttack.java
@@ -30,7 +_,7 @@
                             && body.isWithinMeleeAttackRange(target)
                             && i.<NearestVisibleLivingEntities>get(nearestEntities).contains(target)) {
                             lookTarget.set(new EntityTracker(target, true));
-                            body.swingForAttack(InteractionHand.MAIN_HAND);
+                            body.swing(InteractionHand.MAIN_HAND);
                             body.doHurtTarget(level, target);
                             attackCoolingDown.setWithExpiry(true, cooldownBetweenAttacks);
                             return true;
