--- a/net/minecraft/world/entity/ai/behavior/RamTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/RamTarget.java
@@ -85,6 +_,8 @@
             float f = 0.25F * (float)(i - j);
             float f1 = Mth.clamp(p_217367_.getSpeed() * 1.65F, 0.2F, 3.0F) + f;
             float f2 = livingentity.isDamageSourceBlocked(p_217366_.damageSources().mobAttack(p_217367_)) ? 0.5F : 1.0F;
+            livingentity.knockback$attacker = p_217367_;
+            livingentity.knockback$cause = io.papermc.paper.event.entity.EntityKnockbackEvent.Cause.ENTITY_ATTACK;
             livingentity.knockback((double)(f2 * f1) * this.getKnockbackForce.applyAsDouble(p_217367_), this.ramDirection.x(), this.ramDirection.z());
             this.finishRam(p_217366_, p_217367_);
             p_217366_.playSound(null, p_217367_, this.getImpactSound.apply(p_217367_), SoundSource.NEUTRAL, 1.0F, 1.0F);
