--- a/net/minecraft/world/entity/ai/behavior/ResetProfession.java
+++ b/net/minecraft/world/entity/ai/behavior/ResetProfession.java
@@ -5,6 +_,9 @@
 import net.minecraft.world.entity.npc.villager.Villager;
 import net.minecraft.world.entity.npc.villager.VillagerData;
 import net.minecraft.world.entity.npc.villager.VillagerProfession;
+import org.bukkit.craftbukkit.entity.CraftVillager;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerCareerChangeEvent;
 
 public class ResetProfession {
     public static BehaviorControl<Villager> create() {
@@ -12,7 +_,14 @@
             VillagerData bodyData = body.getVillagerData();
             boolean canBeFired = !bodyData.profession().is(VillagerProfession.NONE) && !bodyData.profession().is(VillagerProfession.NITWIT);
             if (canBeFired && body.getVillagerXp() == 0 && bodyData.level() <= 1) {
-                body.setVillagerData(body.getVillagerData().withProfession(level.registryAccess(), VillagerProfession.NONE));
+                // CraftBukkit start
+                VillagerCareerChangeEvent event = CraftEventFactory.callVillagerCareerChangeEvent(body, CraftVillager.CraftProfession.minecraftHolderToBukkit(level.registryAccess().getOrThrow(VillagerProfession.NONE)), VillagerCareerChangeEvent.ChangeReason.LOSING_JOB);
+                if (event.isCancelled()) {
+                    return false;
+                }
+
+                body.setVillagerData(body.getVillagerData().withProfession(CraftVillager.CraftProfession.bukkitToMinecraftHolder(event.getProfession())));
+                // CraftBukkit end
                 body.refreshBrain(level);
                 return true;
             } else {
