--- a/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
+++ b/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
@@ -6,6 +_,9 @@
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.ai.behavior.declarative.BehaviorBuilder;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class StopAttackingIfTargetInvalid {
     private static final int TIMEOUT_TO_GET_WITHIN_ATTACK_RANGE = 200;
@@ -40,6 +_,17 @@
                             && !stopAttackingWhen.test(level, target)) {
                             return true;
                         } else {
+                            // CraftBukkit start
+                            LivingEntity old = body.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).orElse(null);
+                            EntityTargetEvent event = CraftEventFactory.callEntityTargetLivingEvent(body, null, (old != null && !old.isAlive()) ? EntityTargetEvent.TargetReason.TARGET_DIED : EntityTargetEvent.TargetReason.FORGOT_TARGET);
+                            if (event.isCancelled()) {
+                                return false;
+                            }
+                            if (event.getTarget() != null) {
+                                body.getBrain().setMemory(MemoryModuleType.ATTACK_TARGET, ((CraftLivingEntity) event.getTarget()).getHandle());
+                                return true;
+                            }
+                            // CraftBukkit end
                             onTargetErased.accept(level, body, target);
                             attackTarget.erase();
                             return true;
