--- a/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
+++ b/net/minecraft/world/entity/ai/behavior/StopAttackingIfTargetInvalid.java
@@ -3,12 +_,13 @@
 import java.util.Optional;
 import java.util.function.BiConsumer;
 import java.util.function.Predicate;
-import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.ai.behavior.declarative.BehaviorBuilder;
-import net.minecraft.world.entity.ai.behavior.declarative.MemoryAccessor;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class StopAttackingIfTargetInvalid {
     private static final int TIMEOUT_TO_GET_WITHIN_ATTACK_RANGE = 200;
@@ -41,6 +_,31 @@
                                     && !p_260357_.test(livingentity)) {
                                     return true;
                                 } else {
+                                    // CraftBukkit start
+                                    // Paper start - better track target change reason
+                                    final EntityTargetEvent.TargetReason reason;
+                                    if (!p_258796_.canAttack(livingentity)) {
+                                        reason = EntityTargetEvent.TargetReason.TARGET_INVALID;
+                                    } else if (p_260319_ && StopAttackingIfTargetInvalid.isTiredOfTryingToReachTarget(p_258796_, p_258801_.tryGet(p_258788_))) {
+                                        reason = EntityTargetEvent.TargetReason.FORGOT_TARGET;
+                                    } else if (!livingentity.isAlive()) {
+                                        reason = EntityTargetEvent.TargetReason.TARGET_DIED;
+                                    } else if (livingentity.level() != p_258796_.level()) {
+                                        reason = EntityTargetEvent.TargetReason.TARGET_OTHER_LEVEL;
+                                    } else {
+                                        reason = EntityTargetEvent.TargetReason.TARGET_INVALID;
+                                    }
+                                    // Paper end
+                                    LivingEntity old = livingentity.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).orElse(null);
+                                    org.bukkit.event.entity.EntityTargetEvent event = CraftEventFactory.callEntityTargetLivingEvent(livingentity, null,  reason); // Paper
+                                    if (event.isCancelled()) {
+                                        return false;
+                                    }
+                                    if (event.getTarget() != null) {
+                                        livingentity.getBrain().setMemory(MemoryModuleType.ATTACK_TARGET, ((CraftLivingEntity) event.getTarget()).getHandle());
+                                        return true;
+                                    }
+                                    // CraftBukkit end
                                     p_259568_.accept(p_258796_, livingentity);
                                     p_258787_.erase();
                                     return true;
