--- a/net/minecraft/world/entity/ai/behavior/Swim.java
+++ b/net/minecraft/world/entity/ai/behavior/Swim.java
@@ -3,26 +_,22 @@
 import com.google.common.collect.ImmutableMap;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.FluidTags;
-import net.minecraft.tags.TagKey;
 import net.minecraft.world.entity.Mob;
-import net.minecraft.world.level.material.Fluid;
 
 public class Swim<T extends Mob> extends Behavior<T> {
     private final float chance;
-    private final TagKey<Fluid> fluid;
 
     public Swim(float chance) {
-        this(chance, FluidTags.ENTITY_FLOATABLE);
-    }
-
-    public Swim(float chance, TagKey<Fluid> fluid) {
         super(ImmutableMap.of());
         this.chance = chance;
-        this.fluid = fluid;
+    }
+
+    public static <T extends Mob> boolean shouldSwim(T mob) {
+        return mob.isInWater() && mob.getFluidHeight(FluidTags.WATER) > mob.getFluidJumpThreshold() || mob.isInLava();
     }
 
     protected boolean checkExtraStartConditions(ServerLevel level, Mob body) {
-        return body.isInFluidDeeperThan(body.getFluidJumpThreshold(), this.fluid) || body.isInLava();
+        return shouldSwim(body);
     }
 
     protected boolean canStillUse(ServerLevel level, Mob body, long timestamp) {
