--- a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
+++ b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
@@ -3,9 +_,7 @@
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
 import java.util.Set;
-import java.util.function.Predicate;
 import java.util.stream.Collectors;
-import net.minecraft.core.component.DataComponents;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.SimpleContainer;
 import net.minecraft.world.entity.EntityTypes;
@@ -50,15 +_,21 @@
             body.gossip(level, target, timestamp);
             boolean isFarmer = body.getVillagerData().profession().is(VillagerProfession.FARMER);
             if (body.hasExcessFood() && (isFarmer || target.wantsMoreFood())) {
-                throwHalfStack(body, itemStack -> itemStack.has(DataComponents.VILLAGER_FOOD), target);
+                throwHalfStack(body, Villager.FOOD_POINTS.keySet(), target);
             }
 
             if (isFarmer && body.getInventory().countItem(Items.WHEAT) > Items.WHEAT.getDefaultMaxStackSize() / 2) {
-                throwHalfStack(body, itemStack -> itemStack.is(Items.WHEAT) && itemStack.has(DataComponents.VILLAGER_FOOD), target);
-            }
+                throwHalfStack(body, ImmutableSet.of(Items.WHEAT), target);
+            }
+
+            // Purpur start - Option for Villager Clerics to farm Nether Wart
+            if (level.purpurConfig.villagerClericsFarmWarts && level.purpurConfig.villagerClericFarmersThrowWarts && body.getVillagerData().profession().is(VillagerProfession.CLERIC) && body.getInventory().countItem(Items.NETHER_WART) > Items.NETHER_WART.getDefaultMaxStackSize() / 2) {
+                throwHalfStack(body, ImmutableSet.of(Items.NETHER_WART), target);
+            }
+            // Purpur end - Option for Villager Clerics to farm Nether Wart
 
             if (!this.trades.isEmpty() && body.getInventory().hasAnyOf(this.trades)) {
-                throwHalfStack(body, itemStack -> this.trades.contains(itemStack.getItem()), target);
+                throwHalfStack(body, this.trades, target);
             }
         }
     }
@@ -73,25 +_,29 @@
         return targetItems.stream().filter(entry -> !selfItems.contains(entry)).collect(Collectors.toSet());
     }
 
-    private static void throwHalfStack(Villager villager, Predicate<ItemStack> predicate, LivingEntity target) {
+    private static void throwHalfStack(Villager villager, Set<Item> items, LivingEntity target) {
         SimpleContainer inventory = villager.getInventory();
         ItemStack toThrow = ItemStack.EMPTY;
         int i = 0;
 
         while (i < inventory.getContainerSize()) {
             ItemStack itemStack;
+            Item item;
             int count;
             label28: {
                 itemStack = inventory.getItem(i);
-                if (!itemStack.isEmpty() && predicate.test(itemStack)) {
-                    if (itemStack.getCount() > itemStack.getMaxStackSize() / 2) {
-                        count = itemStack.getCount() / 2;
-                        break label28;
-                    }
+                if (!itemStack.isEmpty()) {
+                    item = itemStack.getItem();
+                    if (items.contains(item)) {
+                        if (itemStack.getCount() > itemStack.getMaxStackSize() / 2) {
+                            count = itemStack.getCount() / 2;
+                            break label28;
+                        }
 
-                    if (itemStack.getCount() > 24) {
-                        count = itemStack.getCount() - 24;
-                        break label28;
+                        if (itemStack.getCount() > 24) {
+                            count = itemStack.getCount() - 24;
+                            break label28;
+                        }
                     }
                 }
 
@@ -99,7 +_,8 @@
                 continue;
             }
 
-            toThrow = itemStack.split(count);
+            itemStack.shrink(count);
+            toThrow = new ItemStack(item, count);
             break;
         }
 
