--- a/net/minecraft/world/entity/ai/behavior/TransportItemsBetweenContainers.java
+++ b/net/minecraft/world/entity/ai/behavior/TransportItemsBetweenContainers.java
@@ -36,6 +_,8 @@
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang3.function.TriConsumer;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.jspecify.annotations.Nullable;
 
 public class TransportItemsBetweenContainers extends Behavior<PathfinderMob> {
@@ -324,7 +_,10 @@
 
         boolean isValidTarget = this.isWantedBlock(body, transportItemTarget.state)
             && !this.isPositionAlreadyVisited(visitedPositions, unreachablePositions, transportItemTarget, level)
-            && !this.isContainerLocked(transportItemTarget);
+            // Paper start - ItemTransportingEntityValidateTargetEvent
+            && !this.isContainerLocked(transportItemTarget)
+            && org.bukkit.craftbukkit.event.CraftEventFactory.callTransporterValidateTarget(body, level, transportItemTarget.pos);
+            // Paper end - ItemTransportingEntityValidateTargetEvent
         return isValidTarget ? transportItemTarget : null;
     }
 
