--- a/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
@@ -39,7 +_,7 @@
             Pair.of(1, new MoveToTargetSink()),
             Pair.of(2, PoiCompetitorScan.create()),
             Pair.of(3, new LookAndFollowTradingPlayerSink(p_24587_)),
-            Pair.of(5, GoToWantedItem.create(p_24587_, false, 4)),
+            Pair.of(5, GoToWantedItem.create(villager -> !villager.isSleeping(), p_24587_, false, 4)),  // Paper - Fix MC-157464
             Pair.of(6, AcquirePoi.create(p_24586_.acquirableJobSite(), MemoryModuleType.JOB_SITE, MemoryModuleType.POTENTIAL_JOB_SITE, true, Optional.empty())),
             Pair.of(7, new GoToPotentialJobSite(p_24587_)),
             Pair.of(8, YieldJobSite.create(p_24587_)),
