--- a/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
@@ -15,6 +_,8 @@
 import net.minecraft.world.entity.ai.village.poi.PoiTypes;
 import net.minecraft.world.entity.npc.villager.Villager;
 import net.minecraft.world.level.pathfinder.Path;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class VillagerMakeLove extends Behavior<Villager> {
     private long birthTimestamp;
@@ -103,12 +_,19 @@
         if (child == null) {
             return Optional.empty();
         }
-
-        source.setAge(6000);
-        target.setAge(6000);
-        child.setAge(-24000);
-        child.snapTo(source.getX(), source.getY(), source.getZ(), 0.0F, 0.0F);
+            child.setAge(-24000);
+            child.snapTo(source.getX(), source.getY(), source.getZ(), 0.0F, 0.0F);
+            // CraftBukkit start - call EntityBreedEvent
+            if (CraftEventFactory.callEntityBreedEvent(child, source, target, null, null, 0).isCancelled()) {
+                return Optional.empty();
+            }
+            // Move age setting down
+            source.setAge(level.purpurConfig.villagerBreedingTicks);
+            target.setAge(level.purpurConfig.villagerBreedingTicks);
+            child.spawnReason(CreatureSpawnEvent.SpawnReason.BREEDING);
         level.addFreshEntityWithPassengers(child);
+        // Neo: If villager is blocked from spawning (e.g., FinalizeSpawnEvent), then breed should be unsuccessful
+        if (!child.isAddedToLevel()) return Optional.empty();
         level.broadcastEntityEvent(child, (byte)12);
         return Optional.of(child);
     }
