--- a/net/minecraft/world/entity/ai/targeting/TargetingConditions.java
+++ b/net/minecraft/world/entity/ai/targeting/TargetingConditions.java
@@ -64,6 +_,10 @@
             return false;
         } else if (this.selector != null && !this.selector.test(p_26887_)) {
             return false;
+            // Purpur start
+        } else if (!p_26887_.level().purpurConfig.idleTimeoutTargetPlayer && p_26887_ instanceof net.minecraft.server.level.ServerPlayer player && player.isAfk()) {
+            return false;
+            // Purpur end
         } else {
             if (p_26886_ == null) {
                 if (this.isCombat && (!p_26887_.canBeSeenAsEnemy() || p_26887_.level().getDifficulty() == Difficulty.PEACEFUL)) {
