--- a/net/minecraft/world/entity/ai/targeting/TargetingConditions.java
+++ b/net/minecraft/world/entity/ai/targeting/TargetingConditions.java
@@ -67,6 +_,10 @@
 
         if (this.selector != null && !this.selector.test(target, level)) {
             return false;
+            // Purpur start - AFK API
+        } else if (!level.purpurConfig.idleTimeoutTargetPlayer && target instanceof net.minecraft.server.level.ServerPlayer player && player.isAfk()) {
+            return false;
+            // Purpur end - AFK API
         }
 
         if (targeter == null) {
@@ -79,7 +_,7 @@
             }
 
             if (this.range > 0.0) {
-                double modifier = this.testInvisible ? target.getVisibilityPercent(level, targeter) : 1.0;
+                double modifier = this.testInvisible ? target.getVisibilityPercent(targeter) : 1.0;
                 double visibilityDistance = Math.max(this.range * modifier, 2.0);
                 double distanceToSqr = targeter.distanceToSqr(target.getX(), target.getY(), target.getZ());
                 if (distanceToSqr > visibilityDistance * visibilityDistance) {
