--- a/com/mojang/realmsclient/RealmsMainScreen.java
+++ b/com/mojang/realmsclient/RealmsMainScreen.java
@@ -12,16 +_,16 @@
 import com.mojang.realmsclient.dto.RegionPingResult;
 import com.mojang.realmsclient.exception.RealmsServiceException;
 import com.mojang.realmsclient.gui.RealmsDataFetcher;
-import com.mojang.realmsclient.gui.RealmsHeader;
 import com.mojang.realmsclient.gui.RealmsServerList;
+import com.mojang.realmsclient.gui.screens.AddRealmPopupScreen;
 import com.mojang.realmsclient.gui.screens.RealmsCreateRealmScreen;
 import com.mojang.realmsclient.gui.screens.RealmsGenericErrorScreen;
-import com.mojang.realmsclient.gui.screens.RealmsJoinRealmWithCodeScreen;
 import com.mojang.realmsclient.gui.screens.RealmsLongRunningMcoTaskScreen;
-import com.mojang.realmsclient.gui.screens.RealmsPdpScreen;
+import com.mojang.realmsclient.gui.screens.RealmsPendingInvitesScreen;
 import com.mojang.realmsclient.gui.screens.RealmsPopups;
 import com.mojang.realmsclient.gui.screens.configuration.RealmsConfigureWorldScreen;
 import com.mojang.realmsclient.gui.task.DataFetcher;
+import com.mojang.realmsclient.util.RealmsPersistence;
 import com.mojang.realmsclient.util.RealmsUtil;
 import com.mojang.realmsclient.util.task.GetServerDetailsTask;
 import java.util.ArrayList;
@@ -51,7 +_,7 @@
 import net.minecraft.client.gui.components.ObjectSelectionList;
 import net.minecraft.client.gui.components.PlayerFaceExtractor;
 import net.minecraft.client.gui.components.PopupScreen;
-import net.minecraft.client.gui.components.RealmsButton;
+import net.minecraft.client.gui.components.SpriteIconButton;
 import net.minecraft.client.gui.components.Tooltip;
 import net.minecraft.client.gui.components.WidgetSprites;
 import net.minecraft.client.gui.components.WidgetTooltipHolder;
@@ -59,12 +_,12 @@
 import net.minecraft.client.gui.layouts.GridLayout;
 import net.minecraft.client.gui.layouts.HeaderAndFooterLayout;
 import net.minecraft.client.gui.layouts.Layout;
+import net.minecraft.client.gui.layouts.LayoutSettings;
 import net.minecraft.client.gui.layouts.LinearLayout;
 import net.minecraft.client.gui.layouts.SpacerElement;
 import net.minecraft.client.gui.navigation.ScreenRectangle;
 import net.minecraft.client.gui.screens.ConfirmLinkScreen;
 import net.minecraft.client.gui.screens.Screen;
-import net.minecraft.client.gui.screens.TitleScreen;
 import net.minecraft.client.gui.screens.inventory.tooltip.ClientActivePlayersTooltip;
 import net.minecraft.client.input.KeyEvent;
 import net.minecraft.client.input.MouseButtonEvent;
@@ -94,6 +_,8 @@
     private static final Identifier EXPIRES_SOON_SPRITE = Identifier.withDefaultNamespace("realm_status/expires_soon");
     private static final Identifier OPEN_SPRITE = Identifier.withDefaultNamespace("realm_status/open");
     private static final Identifier CLOSED_SPRITE = Identifier.withDefaultNamespace("realm_status/closed");
+    private static final Identifier INVITE_SPRITE = Identifier.withDefaultNamespace("icon/invite");
+    private static final Identifier NEWS_SPRITE = Identifier.withDefaultNamespace("icon/news");
     public static final Identifier HARDCORE_MODE_SPRITE = Identifier.withDefaultNamespace("hud/heart/hardcore_full");
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final Identifier NO_REALMS_LOCATION = Identifier.withDefaultNamespace("textures/gui/realms/no_realms.png");
@@ -113,8 +_,8 @@
     private static final Component SERVER_CLOSED_TOOLTIP = Component.translatable("mco.selectServer.closed");
     private static final Component UNITIALIZED_WORLD_NARRATION = Component.translatable("gui.narrate.button", SERVER_UNITIALIZED_TEXT);
     private static final Component NO_REALMS_TEXT = Component.translatable("mco.selectServer.noRealms");
-    private static final Component ADD_REALM_TEXT = Component.translatable("mco.selectServer.purchase");
-    private static final Component FREE_TRIAL = Component.translatable("mco.selectServer.freeTrial");
+    private static final Component NO_PENDING_INVITES = Component.translatable("mco.invites.nopending");
+    private static final Component PENDING_INVITES = Component.translatable("mco.invites.pending");
     private static final Component INCOMPATIBLE_POPUP_TITLE = Component.translatable("mco.compatibility.incompatible.popup.title");
     private static final Component INCOMPATIBLE_RELEASE_TYPE_POPUP_MESSAGE = Component.translatable("mco.compatibility.incompatible.releaseType.popup.message");
     private static final int BUTTON_WIDTH = 100;
@@ -143,19 +_,18 @@
     private RealmsServerList serverList;
     private List<RealmsServer> availableSnapshotServers = List.of();
     private RealmsServerPlayerLists onlinePlayersPerRealm = new RealmsServerPlayerLists(Map.of());
-    private boolean redirectToPdpIfNoRealms;
-    private boolean trialAvailable;
+    private volatile boolean trialsAvailable;
+    private volatile @Nullable String newsLink;
     private final List<RealmsNotification> notifications = new ArrayList<>();
-    private final RealmsHeader header;
     private Button addRealmButton;
+    private RealmsMainScreen.NotificationButton pendingInvitesButton;
+    private RealmsMainScreen.NotificationButton newsButton;
     private RealmsMainScreen.LayoutState activeLayoutState;
     private @Nullable HeaderAndFooterLayout layout;
 
     public RealmsMainScreen(Screen lastScreen) {
         super(TITLE);
         this.lastScreen = lastScreen;
-        this.redirectToPdpIfNoRealms = lastScreen instanceof TitleScreen && Minecraft.getInstance().gui.screen() == lastScreen;
-        this.header = new RealmsHeader(this, this::openJoinRealmScreen);
         this.inviteNarrationLimiter = RateLimiter.create(0.016666668F);
     }
 
@@ -163,12 +_,42 @@
     public void init() {
         this.serverList = new RealmsServerList(this.minecraft);
         this.realmSelectionList = new RealmsMainScreen.RealmSelectionList();
+        Component invitesTitle = Component.translatable("mco.invites.title");
+        this.pendingInvitesButton = new RealmsMainScreen.NotificationButton(
+            invitesTitle, INVITE_SPRITE, var2x -> this.minecraft.gui.setScreen(new RealmsPendingInvitesScreen(this, invitesTitle)), null
+        );
+        Component newsTitle = Component.translatable("mco.news");
+        this.newsButton = new RealmsMainScreen.NotificationButton(newsTitle, NEWS_SPRITE, b -> {
+            String newsLink = this.newsLink;
+            if (newsLink != null) {
+                ConfirmLinkScreen.confirmLinkNow(this, newsLink);
+                if (this.newsButton.notificationCount() != 0) {
+                    RealmsPersistence.RealmsPersistenceData data = RealmsPersistence.readFile();
+                    data.hasUnreadNews = false;
+                    RealmsPersistence.writeFile(data);
+                    this.newsButton.setNotificationCount(0);
+                }
+            }
+        }, newsTitle);
         this.playButton = Button.builder(PLAY_TEXT, button -> play(this.getSelectedServer(), this)).width(100).build();
         this.configureButton = Button.builder(CONFIGURE_SERVER_TEXT, button -> this.configureClicked(this.getSelectedServer())).width(100).build();
         this.renewButton = Button.builder(SUBSCRIPTION_RENEW_TEXT, button -> this.onRenew(this.getSelectedServer())).width(100).build();
         this.leaveButton = Button.builder(LEAVE_SERVER_TEXT, button -> this.leaveClicked(this.getSelectedServer())).width(100).build();
-        this.addRealmButton = this.createAddRealmButton();
+        this.addRealmButton = Button.builder(Component.translatable("mco.selectServer.purchase"), button -> this.openTrialAvailablePopup())
+            .size(100, 20)
+            .build();
         this.backButton = Button.builder(CommonComponents.GUI_BACK, button -> this.onClose()).width(100).build();
+        if (RealmsClient.ENVIRONMENT == RealmsClient.Environment.STAGE) {
+            this.addRenderableWidget(
+                CycleButton.booleanBuilder(Component.literal("Snapshot"), Component.literal("Release"), snapshotToggle)
+                    .create(5, 5, 100, 20, Component.literal("Realm"), (button, value) -> {
+                        snapshotToggle = value;
+                        this.availableSnapshotServers = List.of();
+                        this.debugRefreshDataFetchers();
+                    })
+            );
+        }
+
         this.updateLayout(RealmsMainScreen.LayoutState.LOADING);
         this.updateButtonStates();
         this.availability.thenAcceptAsync(result -> {
@@ -207,11 +_,7 @@
     }
 
     private void updateLayout(RealmsMainScreen.LayoutState state) {
-        this.updateLayout(state, false);
-    }
-
-    private void updateLayout(RealmsMainScreen.LayoutState state, boolean force) {
-        if (force || this.activeLayoutState != state) {
+        if (this.activeLayoutState != state) {
             if (this.layout != null) {
                 this.layout.visitWidgets(x$0 -> this.removeWidget(x$0));
             }
@@ -246,31 +_,22 @@
     }
 
     private Layout createHeader() {
-        return this.header.createLayout(realmsLogo(), 44, 4);
-    }
-
-    private Button createAddRealmButton() {
-        return this.trialAvailable
-            ? new RealmsButton(0, 0, 100, 20, FREE_TRIAL, var1 -> this.openRealmsPdp())
-            : Button.builder(ADD_REALM_TEXT, var1 -> this.openRealmsPdp()).size(100, 20).build();
+        int sideCellWidth = 90;
+        LinearLayout buttons = LinearLayout.horizontal().spacing(4);
+        buttons.defaultCellSetting().alignVerticallyMiddle();
+        buttons.addChild(this.pendingInvitesButton);
+        buttons.addChild(this.newsButton);
+        LinearLayout header = LinearLayout.horizontal();
+        header.defaultCellSetting().alignVerticallyMiddle();
+        header.addChild(SpacerElement.width(90));
+        header.addChild(realmsLogo(), LayoutSettings::alignHorizontallyCenter);
+        header.addChild(new FrameLayout(90, 44)).addChild(buttons, LayoutSettings::alignHorizontallyRight);
+        return header;
     }
 
     private Layout createFooter(RealmsMainScreen.LayoutState state) {
         GridLayout footer = new GridLayout().spacing(4);
         GridLayout.RowHelper helper = footer.createRowHelper(3);
-        if (RealmsClient.ENVIRONMENT == RealmsClient.Environment.STAGE) {
-            helper.addChild(
-                CycleButton.booleanBuilder(Component.literal("Snapshot"), Component.literal("Release"), snapshotToggle)
-                    .create(0, 0, 100, 20, Component.literal("Realm"), (button, value) -> {
-                        snapshotToggle = value;
-                        this.availableSnapshotServers = List.of();
-                        this.debugRefreshDataFetchers();
-                    }),
-                3,
-                helper.newCellSettings().alignHorizontallyCenter()
-            );
-        }
-
         if (state == RealmsMainScreen.LayoutState.LIST) {
             helper.addChild(this.playButton);
             helper.addChild(this.configureButton);
@@ -301,7 +_,6 @@
         RealmsServer server = this.getSelectedServer();
         boolean serverSelected = server != null;
         this.addRealmButton.active = this.activeLayoutState != RealmsMainScreen.LayoutState.LOADING;
-        this.header.setJoinRealmButtonActive(this.activeLayoutState != RealmsMainScreen.LayoutState.LOADING);
         this.playButton.active = serverSelected && server.shouldPlayButtonBeActive();
         if (!this.playButton.active && serverSelected && server.state == RealmsServer.State.CLOSED) {
             this.playButton.setTooltip(Tooltip.create(RealmsServer.WORLD_CLOSED_COMPONENT));
@@ -352,7 +_,6 @@
             this.serverList.updateServersList(updatedServers.serverList());
             this.availableSnapshotServers = updatedServers.availableSnapshotServers();
             this.refreshListAndLayout();
-            this.openPdpIfNoRealms();
             boolean ownsNonExpiredRealmServer = false;
 
             for (RealmsServer retrievedServer : this.serverList) {
@@ -386,16 +_,18 @@
             }
         });
         result.subscribe(dataSource.pendingInvitesTask, numberOfPendingInvites -> {
-            this.header.setPendingInvites(numberOfPendingInvites);
+            this.pendingInvitesButton.setNotificationCount(numberOfPendingInvites);
+            this.pendingInvitesButton.setTooltip(numberOfPendingInvites == 0 ? Tooltip.create(NO_PENDING_INVITES) : Tooltip.create(PENDING_INVITES));
             if (numberOfPendingInvites > 0 && this.inviteNarrationLimiter.tryAcquire(1)) {
                 this.minecraft.getNarrator().saySystemNow(Component.translatable("mco.configure.world.invite.narration", numberOfPendingInvites));
             }
         });
-        result.subscribe(dataSource.trialAvailabilityTask, this::setTrialAvailable);
+        result.subscribe(dataSource.trialAvailabilityTask, newStatus -> this.trialsAvailable = newStatus);
         result.subscribe(dataSource.onlinePlayersTask, playerList -> this.onlinePlayersPerRealm = playerList);
         result.subscribe(dataSource.newsTask, news -> {
             dataSource.newsManager.updateUnreadNews(news);
-            this.header.setNews(dataSource.newsManager.newsLink(), dataSource.newsManager.hasUnreadNews());
+            this.newsLink = dataSource.newsManager.newsLink();
+            this.newsButton.setNotificationCount(dataSource.newsManager.hasUnreadNews() ? Integer.MAX_VALUE : 0);
         });
         return result;
     }
@@ -437,26 +_,6 @@
         this.updateButtonStates();
     }
 
-    private void openPdpIfNoRealms() {
-        if (this.redirectToPdpIfNoRealms) {
-            this.redirectToPdpIfNoRealms = false;
-            if (this.serverList.isEmpty() && this.availableSnapshotServers.isEmpty() && this.minecraft.gui.screen() == this) {
-                this.minecraft
-                    .gui
-                    .setScreen(new RealmsPdpScreen(this.lastScreen, this.trialAvailable, this::openJoinRealmScreen, () -> this.minecraft.gui.setScreen(this)));
-            }
-        }
-    }
-
-    private void setTrialAvailable(boolean trialAvailable) {
-        if (this.trialAvailable != trialAvailable) {
-            this.trialAvailable = trialAvailable;
-            this.addRealmButton = this.createAddRealmButton();
-            this.updateLayout(this.activeLayoutState, true);
-            this.updateButtonStates();
-        }
-    }
-
     private void pingRegions() {
         Util.nonCriticalIoPool().execute(() -> {
             List<RegionPingResult> regionPingResultList = Ping.pingAllRegions();
@@ -563,6 +_,10 @@
             graphics.text(this.font, "Minecraft " + SharedConstants.getCurrentVersion().name(), 2, this.height - 10, -1);
         }
 
+        if (this.trialsAvailable && this.addRealmButton.active) {
+            AddRealmPopupScreen.extractDiamond(graphics, this.addRealmButton);
+        }
+
         switch (RealmsClient.ENVIRONMENT) {
             case STAGE:
                 this.extractEnvironment(graphics, "STAGE!", -256);
@@ -572,18 +_,8 @@
         }
     }
 
-    private void openRealmsPdp() {
-        this.minecraft.gui.setScreen(new RealmsPdpScreen(this, this.trialAvailable, this::openJoinRealmScreen));
-    }
-
-    private void openJoinRealmScreen() {
-        Screen lastScreen = this.minecraft.gui.screen();
-        if (lastScreen != null) {
-            this.minecraft.gui.setScreen(new RealmsJoinRealmWithCodeScreen(lastScreen, () -> {
-                this.minecraft.gui.setScreen(this);
-                this.resetScreen();
-            }));
-        }
+    private void openTrialAvailablePopup() {
+        this.minecraft.gui.setScreen(new AddRealmPopupScreen(this, this.trialsAvailable));
     }
 
     public static void play(@Nullable RealmsServer server, Screen cancelScreen) {
@@ -940,6 +_,52 @@
         LOADING,
         NO_REALMS,
         LIST;
+    }
+
+    private static class NotificationButton extends SpriteIconButton.CenteredIcon {
+        private static final Identifier[] NOTIFICATION_ICONS = new Identifier[]{
+            Identifier.withDefaultNamespace("notification/1"),
+            Identifier.withDefaultNamespace("notification/2"),
+            Identifier.withDefaultNamespace("notification/3"),
+            Identifier.withDefaultNamespace("notification/4"),
+            Identifier.withDefaultNamespace("notification/5"),
+            Identifier.withDefaultNamespace("notification/more")
+        };
+        private static final int UNKNOWN_COUNT = Integer.MAX_VALUE;
+        private static final int SIZE = 20;
+        private static final int SPRITE_SIZE = 14;
+        private int notificationCount;
+
+        public NotificationButton(Component title, Identifier texture, Button.OnPress onPress, @Nullable Component tooltip) {
+            super(20, 20, title, 14, 14, 0, 0, new WidgetSprites(texture), onPress, tooltip, null, false);
+        }
+
+        private int notificationCount() {
+            return this.notificationCount;
+        }
+
+        public void setNotificationCount(int notificationCount) {
+            this.notificationCount = notificationCount;
+        }
+
+        @Override
+        public void extractContents(GuiGraphicsExtractor graphics, int mouseX, int mouseY, float a) {
+            super.extractContents(graphics, mouseX, mouseY, a);
+            if (this.isActive() && this.notificationCount != 0) {
+                this.extractNotificationCounter(graphics);
+            }
+        }
+
+        private void extractNotificationCounter(GuiGraphicsExtractor graphics) {
+            graphics.blitSprite(
+                RenderPipelines.GUI_TEXTURED,
+                NOTIFICATION_ICONS[Math.min(this.notificationCount, 6) - 1],
+                this.getX() + this.getWidth() - 5,
+                this.getY() - 3,
+                8,
+                8
+            );
+        }
     }
 
     private class NotificationMessageEntry extends RealmsMainScreen.Entry {
