--- a/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
@@ -9,7 +_,6 @@
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.level.gamerules.GameRules;
 
 public class EatBlockGoal extends Goal {
     private static final int EAT_ANIMATION_TICKS = 40;
@@ -60,8 +_,9 @@
         this.eatAnimationTick = Math.max(0, this.eatAnimationTick - 1);
         if (this.eatAnimationTick == this.adjustedTickDelay(4)) {
             BlockPos pos = this.mob.blockPosition();
-            if (IS_EDIBLE.test(this.level.getBlockState(pos))) {
-                if (getServerLevel(this.level).getGameRules().get(GameRules.MOB_GRIEFING)) {
+            final BlockState blockState = this.level.getBlockState(pos); // Paper - fix wrong block state
+            if (IS_EDIBLE.test(blockState)) { // Paper - fix wrong block state
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.mob, pos, blockState.getFluidState().createLegacyBlock(), !net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(this.level), this.mob))) { // CraftBukkit // Paper - fix wrong block state
                     this.level.destroyBlock(pos, false);
                 }
 
@@ -69,7 +_,7 @@
             } else {
                 BlockPos below = pos.below();
                 if (this.level.getBlockState(below).is(Blocks.GRASS_BLOCK)) {
-                    if (getServerLevel(this.level).getGameRules().get(GameRules.MOB_GRIEFING)) {
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.mob, below, Blocks.AIR.defaultBlockState(), !net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(this.level), this.mob))) { // CraftBukkit
                         this.level.levelEvent(2001, below, Block.getId(Blocks.GRASS_BLOCK.defaultBlockState()));
                         this.level.setBlock(below, Blocks.DIRT.defaultBlockState(), 2);
                     }
