--- a/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/MeleeAttackGoal.java
@@ -112,7 +_,7 @@
                     this.ticksUntilNextPathRecalculation += 5;
                 }
 
-                if (!this.mob.getNavigation().moveTo(target, 0, this.speedModifier)) {
+                if (!this.mob.getNavigation().moveTo(target, this.speedModifier)) {
                     this.ticksUntilNextPathRecalculation += 15;
                 }
 
@@ -127,7 +_,7 @@
     protected void checkAndPerformAttack(LivingEntity target) {
         if (this.canPerformAttack(target)) {
             this.resetAttackCooldown();
-            this.mob.swingForAttack(InteractionHand.MAIN_HAND);
+            this.mob.swing(InteractionHand.MAIN_HAND);
             this.mob.doHurtTarget(getServerLevel(this.mob), target);
         }
     }
