--- a/net/minecraft/util/thread/BlockableEventLoop.java
+++ b/net/minecraft/util/thread/BlockableEventLoop.java
@@ -56,10 +_,15 @@
     }
 
     public CompletableFuture<Void> submitAsync(Runnable p_18690_) {
-        return CompletableFuture.supplyAsync(() -> {
+        return CompletableFuture.<Void>supplyAsync(() -> {
             p_18690_.run();
             return null;
-        }, this);
+        }, this)
+        .exceptionallyCompose(ex -> {
+            // Neo: Log since this is usually swallowed
+            LOGGER.error(LogUtils.FATAL_MARKER, "Error executing task on {}", name(), ex);
+            return CompletableFuture.failedStage(ex);
+        });
     }
 
     @CheckReturnValue
@@ -80,6 +_,14 @@
         }
     }
 
+    // Paper start
+    public void scheduleOnMain(Runnable runnable) {
+        // postToMainThread does not work the same as older versions of mc
+        // This method is actually used to create a TickTask, which can then be posted onto main
+        this.tell(this.wrapRunnable(runnable));
+    }
+    // Paper end
+
     public void tell(R p_18712_) {
         this.pendingRunnables.add(p_18712_);
         LockSupport.unpark(this.getRunningThread());
@@ -142,6 +_,7 @@
         try {
             p_18700_.run();
         } catch (Exception exception) {
+            if (exception.getCause() instanceof ThreadDeath) throw exception; // Paper
             LOGGER.error(LogUtils.FATAL_MARKER, "Error executing task on {}", this.name(), exception);
         }
     }
