--- a/net/minecraft/network/syncher/EntityDataSerializers.java
+++ b/net/minecraft/network/syncher/EntityDataSerializers.java
@@ -60,10 +_,27 @@
     public static final EntityDataSerializer<Optional<Component>> OPTIONAL_COMPONENT = EntityDataSerializer.forValueType(
         ComponentSerialization.TRUSTED_OPTIONAL_STREAM_CODEC
     );
+    // Paper start - do not obfuscate items sent as entity data
+    public static final StreamCodec<RegistryFriendlyByteBuf, ItemStack> OVERSIZED_ITEM_CODEC = new net.minecraft.network.codec.StreamCodec<>() {
+        @Override
+        public net.minecraft.world.item.ItemStack decode(final net.minecraft.network.RegistryFriendlyByteBuf buffer) {
+            return ItemStack.OPTIONAL_STREAM_CODEC.decode(buffer);
+        }
+
+        @Override
+        public void encode(final net.minecraft.network.RegistryFriendlyByteBuf buffer, final net.minecraft.world.item.ItemStack value) {
+            // If the codec is called during an obfuscation session, downgrade the context's obf level to OVERSIZED if it isn't already.
+            // Entity data cannot be fully obfuscated as entities might render out specific values (e.g. count or custom name).
+            try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.level(io.papermc.paper.util.sanitizer.ItemObfuscationSession.ObfuscationLevel.OVERSIZED))) {
+                ItemStack.OPTIONAL_STREAM_CODEC.encode(buffer, value);
+            }
+        }
+    };
+    // Paper end - do not obfuscate items sent as entity data
     public static final EntityDataSerializer<ItemStack> ITEM_STACK = new EntityDataSerializer<ItemStack>() {
         @Override
         public StreamCodec<? super RegistryFriendlyByteBuf, ItemStack> codec() {
-            return ItemStack.OPTIONAL_STREAM_CODEC;
+            return OVERSIZED_ITEM_CODEC; // Paper - do not obfuscate items sent as entity data
         }
 
         public ItemStack copy(ItemStack value) {
@@ -147,16 +_,27 @@
     public static final EntityDataSerializer<HumanoidArm> HUMANOID_ARM = EntityDataSerializer.forValueType(HumanoidArm.STREAM_CODEC);
     public static final EntityDataSerializer<DyeColor> DYE_COLOR = EntityDataSerializer.forValueType(DyeColor.STREAM_CODEC);
 
+    private static final org.slf4j.Logger LOGGER = com.mojang.logging.LogUtils.getLogger();
+    private static final StackWalker STACK_WALKER = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE);
+    /**
+     * @deprecated NeoForge: Use {@link net.neoforged.neoforge.registries.NeoForgeRegistries#ENTITY_DATA_SERIALIZERS} instead
+     */
+    @Deprecated
     public static void registerSerializer(EntityDataSerializer<?> serializer) {
-        SERIALIZERS.add(serializer);
+        if (!STACK_WALKER.getCallerClass().equals(EntityDataSerializers.class)) {
+            LOGGER.error("Modded EntityDataSerializers must be registered to NeoForgeRegistries.ENTITY_DATA_SERIALIZERS instead to prevent ID mismatches between client and server!");
+            throw new UnsupportedOperationException("Modded EntityDataSerializers must be registered to NeoForgeRegistries.ENTITY_DATA_SERIALIZERS instead to prevent ID mismatches between client and server!");
+        }
+        if (SERIALIZERS.add(serializer) >= net.neoforged.neoforge.common.CommonHooks.VANILLA_SERIALIZER_LIMIT)
+            throw new RuntimeException("Vanilla EntityDataSerializer ID limit exceeded");
     }
 
     public static @Nullable EntityDataSerializer<?> getSerializer(int id) {
-        return SERIALIZERS.byId(id);
+        return net.neoforged.neoforge.common.CommonHooks.getSerializer(id, SERIALIZERS);
     }
 
     public static int getSerializedId(EntityDataSerializer<?> serializer) {
-        return SERIALIZERS.getId(serializer);
+        return net.neoforged.neoforge.common.CommonHooks.getSerializerId(serializer, SERIALIZERS);
     }
 
     private EntityDataSerializers() {
