--- a/net/minecraft/world/item/crafting/display/SlotDisplayContext.java
+++ b/net/minecraft/world/item/crafting/display/SlotDisplayContext.java
@@ -11,6 +_,6 @@
     public static final ContextKeySet CONTEXT = new ContextKeySet.Builder().optional(REGISTRIES).build();
 
     public static ContextMap fromLevel(Level level) {
-        return ContextMap.builder().set(REGISTRIES, level.registryAccess()).buildAndValidate(CONTEXT);
+        return new ContextMap.Builder().withParameter(REGISTRIES, level.registryAccess()).create(CONTEXT);
     }
 }
