--- a/com/mojang/renderpearl/frontend/FrontendGpuDevice.java
+++ b/com/mojang/renderpearl/frontend/FrontendGpuDevice.java
@@ -1,7 +_,6 @@
 package com.mojang.renderpearl.frontend;
 
 import com.mojang.jtracy.TracyClient;
-import com.mojang.logging.LogUtils;
 import com.mojang.renderpearl.api.GpuFormat;
 import com.mojang.renderpearl.api.buffers.GpuBuffer;
 import com.mojang.renderpearl.api.commands.CommandEncoder;
@@ -18,29 +_,20 @@
 import com.mojang.renderpearl.api.textures.GpuTexture;
 import com.mojang.renderpearl.api.textures.GpuTextureView;
 import com.mojang.renderpearl.backend.api.GpuDeviceBackend;
-import com.mojang.renderpearl.frontend.shaders.PipelineBuilder;
 import java.nio.ByteBuffer;
 import java.util.List;
 import java.util.OptionalDouble;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.Executor;
-import java.util.function.BooleanSupplier;
 import java.util.function.Supplier;
-import net.minecraft.SharedConstants;
 import net.minecraft.util.Mth;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
-import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
 public class FrontendGpuDevice implements GpuDevice {
-    private static final Logger LOGGER = LogUtils.getLogger();
-    public static final boolean STRICT_VALIDATION = SharedConstants.IS_RUNNING_IN_IDE;
     private final GpuDeviceBackend backend;
     private final @Nullable TracyGpuProfiler profiler;
     private final CommandEncoder encoder;
-    private final PipelineBuilder pipelineBuilder;
 
     public FrontendGpuDevice(GpuDeviceBackend backend) {
         this.backend = backend;
@@ -51,12 +_,11 @@
         }
 
         this.encoder = new FrontendCommandEncoder(this.profiler, backend, backend.createCommandEncoder());
-        this.pipelineBuilder = new PipelineBuilder(backend);
     }
 
     @Override
-    public GpuSurface createSurface(long windowHandle, BooleanSupplier isIconified) {
-        return new FrontendGpuSurface(this.backend.createSurface(windowHandle, isIconified));
+    public GpuSurface createSurface(long windowHandle) {
+        return new FrontendGpuSurface(this.backend.createSurface(windowHandle));
     }
 
     @Override
@@ -80,7 +_,8 @@
     public GpuTexture createTexture(
         @Nullable Supplier<String> label, @GpuTexture.Usage int usage, GpuFormat format, int width, int height, int depthOrLayers, int mipLevels
     ) {
-        return this.createTexture(this.isDebuggingEnabled() && label != null ? label.get() : null, usage, format, width, height, depthOrLayers, mipLevels);
+        this.verifyTextureCreationArgs(usage, width, height, depthOrLayers, mipLevels);
+        return this.backend.createTexture(label, usage, format, width, height, depthOrLayers, mipLevels);
     }
 
     @Override
@@ -136,7 +_,8 @@
 
     @Override
     public GpuTextureView createTextureView(GpuTexture texture) {
-        return this.createTextureView(texture, 0, texture.getMipLevels());
+        this.verifyTextureViewCreationArgs(texture, 0, texture.getMipLevels());
+        return this.backend.createTextureView(texture, 0, texture.getMipLevels());
     }
 
     @Override
@@ -189,13 +_,12 @@
     }
 
     @Override
-    public CompletableFuture<CompiledRenderPipeline.Pending> compilePipeline(RenderPipeline pipeline, ShaderSource shaderSource, Executor executor) {
-        return this.pipelineBuilder.compilePipeline(pipeline, shaderSource, executor);
+    public @Nullable CompiledRenderPipeline compilePipeline(RenderPipeline pipeline, ShaderSource shaderSource) {
+        return this.backend.compilePipeline(pipeline, shaderSource);
     }
 
     @Override
     public void close() {
-        this.pipelineBuilder.close();
         this.backend.close();
     }
 
@@ -204,8 +_,8 @@
         return this.backend.createTimestampQueryPool(size);
     }
 
-    long getTimestampCalibrationOffset() {
-        return this.backend.getTimestampCalibrationOffset();
+    long getTimestampNow() {
+        return this.backend.getTimestampNow();
     }
 
     @Override
