--- a/com/mojang/renderpearl/frontend/FrontendRenderPass.java
+++ b/com/mojang/renderpearl/frontend/FrontendRenderPass.java
@@ -7,7 +_,6 @@
 import com.mojang.renderpearl.api.commands.RenderPassDescriptor;
 import com.mojang.renderpearl.api.device.DeviceFeatures;
 import com.mojang.renderpearl.api.device.DeviceLimits;
-import com.mojang.renderpearl.api.pipeline.BindGroupLayout;
 import com.mojang.renderpearl.api.pipeline.ColorTargetState;
 import com.mojang.renderpearl.api.pipeline.CompiledRenderPipeline;
 import com.mojang.renderpearl.api.pipeline.IndexType;
@@ -15,16 +_,10 @@
 import com.mojang.renderpearl.api.textures.GpuTextureView;
 import com.mojang.renderpearl.backend.api.GpuDeviceBackend;
 import com.mojang.renderpearl.backend.api.RenderPassBackend;
-import com.mojang.renderpearl.backend.common.BaseGpuBuffer;
-import com.mojang.renderpearl.util.TextureViewAndSampler;
-import java.nio.ByteBuffer;
 import java.nio.IntBuffer;
 import java.util.Collection;
-import java.util.HashMap;
 import java.util.List;
-import java.util.Locale;
 import java.util.Optional;
-import java.util.function.BiConsumer;
 import java.util.function.Supplier;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
@@ -35,28 +_,21 @@
 import org.lwjgl.vulkan.VkDrawIndirectCommand;
 
 @OnlyIn(Dist.CLIENT)
-public class FrontendRenderPass implements RenderPass, RenderPass.UniformUploader {
+public class FrontendRenderPass implements RenderPass {
     private final RenderPassBackend backend;
     private final GpuDeviceBackend device;
     private final DeviceFeatures deviceFeatures;
     private final DeviceLimits deviceLimits;
     private final Runnable onFinish;
     private final RenderPass.@Nullable RenderArea renderArea;
-    private final List<RenderPassDescriptor.@Nullable Attachment<Optional<Vector4fc>>> colorAttachments;
-    private final boolean hasDepthAttachment;
     private boolean isClosed;
     private int pushedDebugGroups;
-    private @Nullable FrontendRenderPipeline boundPipeline;
-    private final @Nullable GpuBufferSlice[] vertexBuffers = new GpuBufferSlice[16];
-    protected @Nullable GpuBuffer indexBuffer;
-    protected final HashMap<String, Object> uniforms = new HashMap<>();
-    private boolean constantsPushed = false;
+    private final List<RenderPassDescriptor.@Nullable Attachment<Optional<Vector4fc>>> colorAttachments;
 
     public FrontendRenderPass(
         RenderPassBackend backend,
         GpuDeviceBackend device,
         List<RenderPassDescriptor.@Nullable Attachment<Optional<Vector4fc>>> colorAttachments,
-        boolean hasDepthAttachment,
         Runnable onFinish,
         RenderPass.@Nullable RenderArea renderArea
     ) {
@@ -65,7 +_,6 @@
         this.deviceFeatures = device.getDeviceInfo().features();
         this.deviceLimits = device.getDeviceInfo().limits();
         this.colorAttachments = colorAttachments;
-        this.hasDepthAttachment = hasDepthAttachment;
         this.onFinish = onFinish;
         this.renderArea = renderArea;
     }
@@ -105,48 +_,32 @@
 
     @Override
     public void setPipeline(CompiledRenderPipeline pipeline) {
-        if (!(pipeline instanceof FrontendRenderPipeline frontendPipeline)) {
-            throw new IllegalArgumentException("Pipeline must be instance of FrontendCompiledRenderPipeline");
-        } else {
-            List<ColorTargetState> colorTargetStates = frontendPipeline.colorTargetStates();
-            if (colorTargetStates.size() != this.colorAttachments.size()) {
-                throw new IllegalStateException("Render pass color attachment count must match pipeline color target state count.");
-            }
+        ColorTargetState[] colorTargetStates = pipeline.info().getColorTargetStates();
+        if (colorTargetStates.length != this.colorAttachments.size()) {
+            throw new IllegalStateException("Render pass color attachment count must match pipeline color target state count.");
+        }
 
-            for (int i = 0; i < this.colorAttachments.size(); i++) {
-                RenderPassDescriptor.Attachment<Optional<Vector4fc>> attachment = this.colorAttachments.get(i);
-                if (attachment != null) {
-                    ColorTargetState colorTargetState = colorTargetStates.get(i);
-                    if (colorTargetState == null || colorTargetState.format() != attachment.textureView().texture().getFormat()) {
-                        throw new IllegalStateException("Render pass color attachment " + i + " format doesn't match pipeline format.");
-                    }
+        for (int i = 0; i < this.colorAttachments.size(); i++) {
+            RenderPassDescriptor.Attachment<Optional<Vector4fc>> attachment = this.colorAttachments.get(i);
+            if (attachment != null) {
+                ColorTargetState colorTargetState = colorTargetStates[i];
+                if (colorTargetState == null || colorTargetState.format() != attachment.textureView().texture().getFormat()) {
+                    throw new IllegalStateException("Render pass color attachment " + i + " format doesn't match pipeline format.");
                 }
             }
-
-            this.boundPipeline = frontendPipeline;
-            this.backend.setPipeline(frontendPipeline.backendRenderPipeline());
-            this.uniforms.forEach(this::setUniform);
-            this.constantsPushed = false;
         }
+
+        this.backend.setPipeline(pipeline);
     }
 
     @Override
-    public void setUniform(String name, @Nullable GpuTextureView textureView, @Nullable GpuSampler sampler) {
-        if (textureView != null && sampler != null) {
-            TextureViewAndSampler pair = new TextureViewAndSampler(textureView, sampler);
-            this.setUniform(name, pair);
-        } else {
-            if (textureView != null || sampler != null) {
-                throw new IllegalArgumentException("textureView and sampler must both or neither be null");
-            }
-
-            this.setUniform(name, (Object)null);
-        }
+    public void bindTexture(String name, @Nullable GpuTextureView textureView, @Nullable GpuSampler sampler) {
+        this.backend.bindTexture(name, textureView, sampler);
     }
 
     @Override
     public void setUniform(String name, GpuBuffer value) {
-        this.setUniform(name, value.slice());
+        this.backend.setUniform(name, value);
     }
 
     @Override
@@ -156,36 +_,7 @@
             throw new IllegalArgumentException("Uniform buffer offset must be aligned to " + alignment);
         }
 
-        this.setUniform(name, (Object)value);
-    }
-
-    private void setUniform(String name, @Nullable Object value) {
-        if (value == null) {
-            this.uniforms.remove(name);
-        } else {
-            this.uniforms.put(name, value);
-        }
-
-        if (this.boundPipeline != null) {
-            int uniformIndex = this.boundPipeline.uniformIndices().getOrDefault(name, -1);
-            if (uniformIndex != -1) {
-                this.backend.setUniform(uniformIndex, value);
-            }
-        }
-    }
-
-    @Override
-    public void pushConstants(ByteBuffer value) {
-        if (this.boundPipeline == null) {
-            throw new IllegalStateException("Must bind pipeline before pushing constants");
-        }
-
-        if (value.remaining() < this.boundPipeline.pushConstantSize()) {
-            throw new IllegalArgumentException("Not enough values for push constant range");
-        }
-
-        this.constantsPushed = true;
-        this.backend.pushConstants(value);
+        this.backend.setUniform(name, value);
     }
 
     @Override
@@ -194,7 +_,7 @@
             if (x >= this.renderArea.x()
                 && y >= this.renderArea.y()
                 && x + width <= this.renderArea.x() + this.renderArea.width()
-                && y + height <= this.renderArea.y() + this.renderArea.height()) {
+                && y + height <= this.renderArea.height()) {
                 this.backend.enableScissor(x, y, width, height);
             } else {
                 throw new IllegalArgumentException(
@@ -225,13 +_,11 @@
             throw new IllegalStateException("Vertex buffer at slot " + slot + " doesn't have GpuBuffer.USAGE_VERTEX flag!");
         }
 
-        this.vertexBuffers[slot] = vertexBuffer;
         this.backend.setVertexBuffer(slot, vertexBuffer);
     }
 
     @Override
     public void setIndexBuffer(GpuBuffer indexBuffer, IndexType indexType) {
-        this.indexBuffer = indexBuffer;
         this.backend.setIndexBuffer(indexBuffer, indexType);
     }
 
@@ -245,7 +_,6 @@
             throw new UnsupportedOperationException("firstInstance must be zero on when device does not support nonZeroFirstInstance");
         }
 
-        this.validateDraw(List.of(), false, true);
         this.backend.drawIndexed(indexCount, instanceCount, firstIndex, vertexOffset, firstInstance);
     }
 
@@ -271,7 +_,6 @@
             throw new IllegalArgumentException("Not enough elements in drawParameters for drawCount draws");
         }
 
-        this.validateDraw(List.of(), false, true);
         this.backend.multiDrawIndexed(drawParameters, instanceCount, firstInstance, drawCount);
     }
 
@@ -297,7 +_,6 @@
             throw new IllegalArgumentException("vertexOffsets does not contain enough elements for drawCount draws");
         }
 
-        this.validateDraw(List.of(), false, true);
         this.backend.multiDrawIndexed(firstIndexOffsets, indexCounts, vertexOffsets, drawCount);
     }
 
@@ -327,7 +_,6 @@
             throw new IllegalArgumentException("Commands offset must be multiple of 4");
         }
 
-        this.validateDraw(List.of(), false, true);
         this.backend.drawIndexedIndirect(commands, drawCount);
     }
 
@@ -343,61 +_,7 @@
             throw new IllegalStateException("Can't use a closed render pass");
         }
 
-        this.validateDraw(dynamicUniforms, true, false);
-        GpuBuffer lastIndexBuffer = null;
-        IndexType lastIndexType = null;
-
-        for (RenderPass.Draw<T> draw : draws) {
-            BiConsumer<T, RenderPass.UniformUploader> uniformUploaderConsumer = draw.uniformUploaderConsumer();
-            if (uniformUploaderConsumer != null) {
-                uniformUploaderConsumer.accept(uniformArgument, this);
-            }
-
-            IndexType indexType = draw.indexType() == null ? defaultIndexType : draw.indexType();
-            GpuBuffer indexBuffer = draw.indexBuffer() == null ? defaultIndexBuffer : draw.indexBuffer();
-            assert indexBuffer != null;
-            assert indexType != null;
-            this.setVertexBuffer(draw.slot(), draw.vertexBuffer().slice());
-            if (FrontendGpuDevice.STRICT_VALIDATION) {
-                if (indexBuffer == null) {
-                    throw new IllegalStateException("Missing index buffer");
-                }
-
-                ((BaseGpuBuffer)indexBuffer).checkCanBeUsed();
-                if (indexBuffer.isClosed()) {
-                    throw new IllegalStateException("Index buffer has been closed!");
-                }
-
-                if (draw.slot() < 0 || draw.slot() >= 16) {
-                    throw new IllegalStateException("Vertex buffer slot must be between 0 and 16");
-                }
-
-                if (this.vertexBuffers[draw.slot()] != null) {
-                    ((BaseGpuBuffer)this.vertexBuffers[draw.slot()].buffer()).checkCanBeUsed();
-                }
-
-                if (this.vertexBuffers[draw.slot()] == null) {
-                    throw new IllegalStateException("Missing vertex buffer at slot " + draw.slot());
-                }
-
-                if (this.vertexBuffers[draw.slot()].buffer().isClosed()) {
-                    throw new IllegalStateException("Vertex buffer at slot " + draw.slot() + " has been closed!");
-                }
-            }
-
-            if (indexBuffer != lastIndexBuffer || indexType != lastIndexType) {
-                this.setIndexBuffer(indexBuffer, indexType);
-                lastIndexBuffer = indexBuffer;
-                lastIndexType = indexType;
-            }
-
-            assert this.boundPipeline != null;
-            if (this.boundPipeline.pushConstantSize() > 0 && !this.constantsPushed) {
-                throw new IllegalStateException("Missing push constants");
-            }
-
-            this.backend.drawIndexed(draw.indexCount(), 1, draw.firstIndex(), draw.baseVertex(), 0);
-        }
+        this.backend.drawMultipleIndexed(draws, defaultIndexBuffer, defaultIndexType, dynamicUniforms, uniformArgument);
     }
 
     @Override
@@ -410,7 +_,6 @@
             throw new UnsupportedOperationException("firstInstance must be zero on when device does not support nonZeroFirstInstance");
         }
 
-        this.validateDraw(List.of(), false, false);
         this.backend.draw(vertexCount, instanceCount, firstVertex, firstInstance);
     }
 
@@ -436,7 +_,6 @@
             throw new IllegalArgumentException("Not enough elements in drawParameters for drawCount draws");
         }
 
-        this.validateDraw(List.of(), false, false);
         this.backend.multiDraw(drawParameters, instanceCount, firstInstance, drawCount);
     }
 
@@ -458,7 +_,6 @@
             throw new IllegalArgumentException("vertexCounts does not contain enough elements for drawCount draws");
         }
 
-        this.validateDraw(List.of(), false, false);
         this.backend.multiDraw(firstVertices, vertexCounts, drawCount);
     }
 
@@ -488,7 +_,6 @@
             throw new IllegalArgumentException("Commands offset must be multiple of 4");
         }
 
-        this.validateDraw(List.of(), false, false);
         this.backend.drawIndirect(commands, drawCount);
     }
 
@@ -501,118 +_,6 @@
             }
 
             this.onFinish.run();
-        }
-    }
-
-    private void validateDraw(Collection<String> dynamicUniforms, boolean dynamicVertexBuffer, boolean indexed) {
-        if (this.boundPipeline == null) {
-            throw new IllegalStateException("Can't draw without a render pipeline");
-        }
-
-        if (FrontendGpuDevice.STRICT_VALIDATION) {
-            if (this.boundPipeline.pushConstantSize() > 0 && !this.constantsPushed) {
-                throw new IllegalStateException("Missing push constants");
-            }
-
-            if (indexed) {
-                if (this.indexBuffer == null) {
-                    throw new IllegalStateException("Missing index buffer");
-                }
-
-                ((BaseGpuBuffer)this.indexBuffer).checkCanBeUsed();
-                if (this.indexBuffer.isClosed()) {
-                    throw new IllegalStateException("Index buffer has been closed!");
-                }
-
-                if ((this.indexBuffer.usage() & 64) == 0) {
-                    throw new IllegalStateException("Index buffer must have GpuBuffer.USAGE_INDEX!");
-                }
-            }
-
-            if (!dynamicVertexBuffer) {
-                for (int i = 0; i < 16; i++) {
-                    if (this.vertexBuffers[i] == null && this.boundPipeline.vertexFormats().get(i) != null) {
-                        throw new IllegalStateException("Vertex format contains elements but vertex buffer at slot " + i + " is null");
-                    }
-
-                    if (this.vertexBuffers[i] != null) {
-                        ((BaseGpuBuffer)this.vertexBuffers[i].buffer()).checkCanBeUsed();
-                    }
-                }
-            }
-
-            for (BindGroupLayout.UniformDescription uniform : this.boundPipeline.uniforms()) {
-                Object value = this.uniforms.get(uniform.name());
-                if (!dynamicUniforms.contains(uniform.name())) {
-                    if (value == null) {
-                        throw new IllegalStateException("Missing uniform " + uniform.name() + " (should be " + uniform.type() + ")");
-                    }
-
-                    switch (uniform.type()) {
-                        case UNIFORM_BUFFER:
-                            if (!(value instanceof GpuBufferSlice valueSlice)) {
-                                throw new IllegalArgumentException("UBO value must be GpuBufferSlice");
-                            }
-
-                            ((BaseGpuBuffer)valueSlice.buffer()).checkCanBeUsed();
-                            if (valueSlice.buffer().isClosed()) {
-                                throw new IllegalStateException("Uniform buffer " + uniform.name() + " is already closed");
-                            }
-
-                            if ((valueSlice.buffer().usage() & 128) == 0) {
-                                throw new IllegalStateException("Uniform buffer " + uniform.name() + " must have GpuBuffer.USAGE_UNIFORM");
-                            }
-                            break;
-                        case TEXEL_BUFFER:
-                            if (value instanceof GpuBufferSlice(GpuBuffer buffer, long offset, long length)) {
-                                if (offset == 0L && length == buffer.size()) {
-                                    if ((buffer.usage() & 256) == 0) {
-                                        throw new IllegalStateException(
-                                            "Uniform texel buffer " + uniform.name() + " must have GpuBuffer.USAGE_UNIFORM_TEXEL_BUFFER"
-                                        );
-                                    }
-
-                                    if (uniform.gpuFormat() == null) {
-                                        throw new IllegalStateException("Invalid uniform texel buffer " + uniform.name() + " (missing a texture format)");
-                                    }
-                                    break;
-                                }
-
-                                throw new IllegalStateException("Uniform texel buffers do not support a slice of a buffer, must be entire buffer");
-                            }
-
-                            throw new IllegalArgumentException("UTB value must be GpuBufferSlice");
-                        case COMBINED_IMAGE_SAMPLER:
-                            if (!(value instanceof TextureViewAndSampler(GpuTextureView textureView, GpuSampler sampler))) {
-                                throw new IllegalArgumentException("Sampler value must be TextureViewAndSampler");
-                            }
-
-                            if (textureView.isClosed()) {
-                                throw new IllegalStateException(
-                                    "Texture view " + uniform.name() + " (" + textureView.texture().getLabel() + ") has been closed!"
-                                );
-                            }
-
-                            if ((textureView.texture().usage() & 4) == 0) {
-                                throw new IllegalStateException(
-                                    "Texture view " + uniform.name() + " (" + textureView.texture().getLabel() + ") must have USAGE_TEXTURE_BINDING!"
-                                );
-                            }
-
-                            if (sampler.isClosed()) {
-                                throw new IllegalStateException(
-                                    "Sampler for " + uniform.name() + " (" + textureView.texture().getLabel() + ") has been closed!"
-                                );
-                            }
-                    }
-                }
-            }
-
-            if (this.boundPipeline.wantsDepthTexture() && !this.hasDepthAttachment) {
-                throw new IllegalStateException(
-                    String.format(Locale.ROOT, "Render pipeline %s wants a depth texture but none was provided", this.boundPipeline.name())
-                );
-            }
         }
     }
 }
