--- a/net/minecraft/client/model/monster/piglin/PiglinModel.java
+++ b/net/minecraft/client/model/monster/piglin/PiglinModel.java
@@ -18,7 +_,7 @@
     public void setupAnim(PiglinRenderState state) {
         super.setupAnim(state);
         float defaultAngle = (float) (Math.PI / 6);
-        float swingAnimation = state.swingAnimation;
+        float attackTime = state.attackTime;
         PiglinArmPose pose = state.armPose;
         if (pose == PiglinArmPose.DANCING) {
             float dancePos = state.ageInTicks / 60.0F;
@@ -31,7 +_,7 @@
             this.rightArm.y = this.rightArm.y + (Mth.sin(dancePos * 40.0F) * 0.5F - 0.5F);
             this.leftArm.y = this.leftArm.y + (Mth.sin(dancePos * 40.0F) * 0.5F + 0.5F);
             this.body.y = this.body.y + Mth.sin(dancePos * 40.0F) * 0.35F;
-        } else if (pose == PiglinArmPose.ATTACKING_WITH_MELEE_WEAPON && swingAnimation == 0.0F) {
+        } else if (pose == PiglinArmPose.ATTACKING_WITH_MELEE_WEAPON && attackTime == 0.0F) {
             this.holdWeaponHigh(state);
         } else if (pose == PiglinArmPose.CROSSBOW_HOLD) {
             AnimationUtils.animateCrossbowHold(this.rightArm, this.leftArm, this.head, state.mainArm == HumanoidArm.RIGHT);
@@ -53,9 +_,9 @@
     }
 
     protected void setupAttackAnimation(PiglinRenderState state) {
-        float swingAnimation = state.swingAnimation;
-        if (swingAnimation > 0.0F && state.armPose == PiglinArmPose.ATTACKING_WITH_MELEE_WEAPON) {
-            AnimationUtils.swingWeaponDown(this.rightArm, this.leftArm, state.mainArm, swingAnimation, state.ageInTicks);
+        float attackTime = state.attackTime;
+        if (attackTime > 0.0F && state.armPose == PiglinArmPose.ATTACKING_WITH_MELEE_WEAPON) {
+            AnimationUtils.swingWeaponDown(this.rightArm, this.leftArm, state.mainArm, attackTime, state.ageInTicks);
         } else {
             super.setupAttackAnimation(state);
         }
