--- a/net/minecraft/client/model/monster/piglin/ZombifiedPiglinModel.java
+++ b/net/minecraft/client/model/monster/piglin/ZombifiedPiglinModel.java
@@ -12,8 +_,8 @@
         super(root);
     }
 
-    protected void setupAttackAnimation(ZombifiedPiglinRenderState state) {
-        super.setupAttackAnimation(state);
+    public void setupAnim(ZombifiedPiglinRenderState state) {
+        super.setupAnim(state);
         AnimationUtils.animateZombieArms(this.leftArm, this.rightArm, state.isAggressive, state);
     }
 }
