--- a/net/minecraft/server/chase/ChaseClient.java
+++ b/net/minecraft/server/chase/ChaseClient.java
@@ -14,6 +_,7 @@
 import java.util.Scanner;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
+import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.commands.ChaseCommand;
@@ -156,6 +_,8 @@
                             Vec2.ZERO,
                             level,
                             LevelBasedPermissionSet.OWNER,
+                            "",
+                            CommonComponents.EMPTY,
                             this.server,
                             player
                         );
