--- a/net/minecraft/advancements/AdvancementHolder.java
+++ b/net/minecraft/advancements/AdvancementHolder.java
@@ -5,6 +_,7 @@
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.resources.Identifier;
+import org.bukkit.craftbukkit.advancement.CraftAdvancement;
 
 public record AdvancementHolder(Identifier id, Advancement value) {
     public static final StreamCodec<RegistryFriendlyByteBuf, AdvancementHolder> STREAM_CODEC = StreamCodec.composite(
@@ -26,4 +_,10 @@
     public String toString() {
         return this.id.toString();
     }
+
+    // CraftBukkit start
+    public final org.bukkit.advancement.Advancement toBukkit() {
+        return new CraftAdvancement(this);
+    }
+    // CraftBukkit end
 }
