--- a/net/minecraft/advancements/Advancement.java
+++ b/net/minecraft/advancements/Advancement.java
@@ -52,6 +_,7 @@
         )
         .validate(Advancement::validate);
     public static final StreamCodec<RegistryFriendlyByteBuf, Advancement> STREAM_CODEC = StreamCodec.ofMember(Advancement::write, Advancement::read);
+    public static final Codec<Optional<net.neoforged.neoforge.common.conditions.WithConditions<Advancement>>> CONDITIONAL_CODEC = net.neoforged.neoforge.common.conditions.ConditionalOps.createConditionalCodecWithConditions(CODEC);
 
     public Advancement(
         Optional<ResourceLocation> p_300893_,
@@ -68,7 +_,7 @@
         return p_312433_.requirements().validate(p_312433_.criteria().keySet()).map(p_311382_ -> p_312433_);
     }
 
-    private static Component decorateName(DisplayInfo p_301019_) {
+    public static Component decorateName(DisplayInfo p_301019_) {
         Component component = p_301019_.getTitle();
         ChatFormatting chatformatting = p_301019_.getType().getChatColor();
         Component component1 = ComponentUtils.mergeStyles(component.copy(), Style.EMPTY.withColor(chatformatting))
