--- a/net/minecraft/advancements/AdvancementProgress.java
+++ b/net/minecraft/advancements/AdvancementProgress.java
@@ -4,7 +_,6 @@
 import com.google.common.collect.Maps;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import io.netty.buffer.ByteBuf;
 import java.time.Instant;
 import java.time.ZoneId;
 import java.time.format.DateTimeFormatter;
@@ -17,9 +_,8 @@
 import java.util.Set;
 import java.util.Map.Entry;
 import java.util.stream.Collectors;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.Component;
-import net.minecraft.network.codec.ByteBufCodecs;
-import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.Util;
 import org.jspecify.annotations.Nullable;
@@ -43,9 +_,6 @@
             )
             .apply(i, (criteria, done) -> new AdvancementProgress(new HashMap<>(criteria)))
     );
-    public static final StreamCodec<ByteBuf, AdvancementProgress> STREAM_CODEC = StreamCodec.composite(
-        ByteBufCodecs.map(HashMap::new, ByteBufCodecs.STRING_UTF8, CriterionProgress.STREAM_CODEC), a -> a.criteria, AdvancementProgress::new
-    );
     private final Map<String, CriterionProgress> criteria;
     private AdvancementRequirements requirements = AdvancementRequirements.EMPTY;
 
@@ -105,6 +_,15 @@
     @Override
     public String toString() {
         return "AdvancementProgress{criteria=" + this.criteria + ", requirements=" + this.requirements + "}";
+    }
+
+    public void serializeToNetwork(FriendlyByteBuf output) {
+        output.writeMap(this.criteria, FriendlyByteBuf::writeUtf, (b, v) -> v.serializeToNetwork(b));
+    }
+
+    public static AdvancementProgress fromNetwork(FriendlyByteBuf input) {
+        Map<String, CriterionProgress> criteria = input.readMap(FriendlyByteBuf::readUtf, CriterionProgress::fromNetwork);
+        return new AdvancementProgress(criteria);
     }
 
     public @Nullable CriterionProgress getCriterion(String id) {
