--- a/net/minecraft/advancements/AdvancementNode.java
+++ b/net/minecraft/advancements/AdvancementNode.java
@@ -9,8 +_,6 @@
     private final AdvancementHolder holder;
     private final @Nullable AdvancementNode parent;
     private final Set<AdvancementNode> children = new ReferenceOpenHashSet<>();
-    private float x;
-    private float y;
 
     @VisibleForTesting
     public AdvancementNode(AdvancementHolder holder, @Nullable AdvancementNode parent) {
@@ -26,14 +_,6 @@
         return this.holder;
     }
 
-    public boolean isTask() {
-        return this.parent != null;
-    }
-
-    public boolean isRoot() {
-        return this.parent == null;
-    }
-
     public @Nullable AdvancementNode parent() {
         return this.parent;
     }
@@ -62,19 +_,6 @@
     @VisibleForTesting
     public void addChild(AdvancementNode child) {
         this.children.add(child);
-    }
-
-    public void setLocation(float x, float y) {
-        this.x = x;
-        this.y = y;
-    }
-
-    public float x() {
-        return this.x;
-    }
-
-    public float y() {
-        return this.y;
     }
 
     @Override
