--- a/net/minecraft/advancements/AdvancementTree.java
+++ b/net/minecraft/advancements/AdvancementTree.java
@@ -1,5 +_,6 @@
 package net.minecraft.advancements;
 
+import com.mohistmc.youer.util.I18n;
 import com.mojang.logging.LogUtils;
 import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
@@ -26,7 +_,7 @@
             this.remove(advancementnode);
         }
 
-        LOGGER.info("Forgot about advancement {}", p_301231_.holder());
+        LOGGER.debug("Forgot about advancement {}", p_301231_.holder());
         this.nodes.remove(p_301231_.holder().id());
         if (p_301231_.parent() == null) {
             this.roots.remove(p_301231_);
@@ -62,7 +_,7 @@
             }
         }
 
-        LOGGER.info("Loaded {} advancements", this.nodes.size());
+        LOGGER.info(I18n.as("advancementtree.loaded", this.nodes.size()));
     }
 
     private boolean tryInsert(AdvancementHolder p_301290_) {
