--- a/net/minecraft/advancements/CriterionProgress.java
+++ b/net/minecraft/advancements/CriterionProgress.java
@@ -1,22 +_,16 @@
 package net.minecraft.advancements;
 
-import io.netty.buffer.ByteBuf;
 import java.time.Instant;
-import java.util.Optional;
-import net.minecraft.network.codec.ByteBufCodecs;
-import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.network.FriendlyByteBuf;
 import org.jspecify.annotations.Nullable;
 
 public class CriterionProgress {
-    public static final StreamCodec<ByteBuf, CriterionProgress> STREAM_CODEC = ByteBufCodecs.INSTANT
-        .apply(ByteBufCodecs::optional)
-        .map(obtained -> new CriterionProgress((Instant)obtained.orElse(null)), progress -> Optional.ofNullable(progress.obtained));
     private @Nullable Instant obtained;
 
     public CriterionProgress() {
     }
 
-    public CriterionProgress(@Nullable Instant obtained) {
+    public CriterionProgress(Instant obtained) {
         this.obtained = obtained;
     }
 
@@ -39,5 +_,15 @@
     @Override
     public String toString() {
         return "CriterionProgress{obtained=" + (this.obtained == null ? "false" : this.obtained) + "}";
+    }
+
+    public void serializeToNetwork(FriendlyByteBuf output) {
+        output.writeNullable(this.obtained, FriendlyByteBuf::writeInstant);
+    }
+
+    public static CriterionProgress fromNetwork(FriendlyByteBuf input) {
+        CriterionProgress result = new CriterionProgress();
+        result.obtained = input.readNullable(FriendlyByteBuf::readInstant);
+        return result;
     }
 }
