--- a/net/minecraft/advancements/TreeNodePosition.java
+++ b/net/minecraft/advancements/TreeNodePosition.java
@@ -192,7 +_,7 @@
     }
 
     private void finalizePosition() {
-        this.node.setLocation(this.x, this.y);
+        this.node.advancement().display().ifPresent(display -> display.setLocation(this.x, this.y));
         if (!this.children.isEmpty()) {
             for (TreeNodePosition child : this.children) {
                 child.finalizePosition();
