--- a/net/minecraft/world/level/levelgen/structure/structures/EndCityPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/EndCityPieces.java
@@ -274,7 +274,12 @@
                 BlockPos blockpos1 = blockpos.below();
 
                 if (boundingbox.isInside(blockpos1)) {
+                    // CraftBukkit start - ensure block transformation
+                    /*
                     RandomizableContainer.setBlockEntityLootTable(serverlevelaccessor, randomsource, blockpos1, BuiltInLootTables.END_CITY_TREASURE);
+                    */
+                    setCraftLootTable(serverlevelaccessor, blockpos1, randomsource, BuiltInLootTables.END_CITY_TREASURE);
+                    // CraftBukkit end
                 }
             } else if (boundingbox.isInside(blockpos) && Level.isInSpawnableBounds(blockpos)) {
                 if (s.startsWith("Sentry")) {
