--- a/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
@@ -86,11 +86,16 @@
         protected void handleDataMarker(String s, BlockPos blockpos, ServerLevelAccessor serverlevelaccessor, RandomSource randomsource, BoundingBox boundingbox) {
             if ("chest".equals(s)) {
                 serverlevelaccessor.setBlock(blockpos, Blocks.AIR.defaultBlockState(), 3);
+                // CraftBukkit start - ensure block transformation
+                /*
                 BlockEntity blockentity = serverlevelaccessor.getBlockEntity(blockpos.below());
 
                 if (blockentity instanceof ChestBlockEntity) {
                     ((ChestBlockEntity) blockentity).setLootTable(BuiltInLootTables.IGLOO_CHEST, randomsource.nextLong());
                 }
+                */
+                setCraftLootTable(serverlevelaccessor, blockpos.below(), randomsource, BuiltInLootTables.IGLOO_CHEST);
+                // CraftBukkit end
 
             }
         }
