--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -516,6 +516,8 @@
 
                         if (boundingbox.isInside(blockpos1) && this.isInterior(worldgenlevel, 1, 0, i2, boundingbox)) {
                             this.hasPlacedSpider = true;
+                            // CraftBukkit start
+                            /*
                             worldgenlevel.setBlock(blockpos1, Blocks.SPAWNER.defaultBlockState(), 2);
                             BlockEntity blockentity = worldgenlevel.getBlockEntity(blockpos1);
 
@@ -524,6 +526,9 @@
 
                                 spawnerblockentity.setEntityId(EntityType.CAVE_SPIDER, randomsource);
                             }
+                            */
+                            placeCraftSpawner(worldgenlevel, blockpos1, org.bukkit.entity.EntityType.CAVE_SPIDER, 2);
+                            // CraftBukkit end
                         }
                     }
                 }
