--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -38,6 +38,10 @@
 import net.minecraft.world.level.storage.loot.LootTable;
 import net.minecraft.world.level.storage.loot.LootTables;
 
+// CraftBukkit start
+import net.minecraft.nbt.NBTBase;
+// CraftBukkit end
+
 public class MineshaftPieces {
 
     private static final int DEFAULT_SHAFT_WIDTH = 3;
@@ -517,6 +521,8 @@
 
                         if (structureboundingbox.isInside(blockposition1) && this.isInterior(generatoraccessseed, 1, 0, i2, structureboundingbox)) {
                             this.hasPlacedSpider = true;
+                            // CraftBukkit start
+                            /*
                             generatoraccessseed.setBlock(blockposition1, Blocks.SPAWNER.defaultBlockState(), 2);
                             TileEntity tileentity = generatoraccessseed.getBlockEntity(blockposition1);
 
@@ -525,6 +531,9 @@
 
                                 tileentitymobspawner.setEntityId(EntityTypes.CAVE_SPIDER, randomsource);
                             }
+                            */
+                            placeCraftSpawner(generatoraccessseed, blockposition1, org.bukkit.entity.EntityType.CAVE_SPIDER, 2);
+                            // CraftBukkit end
                         }
                     }
                 }
