--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -667,6 +667,8 @@
 
                 if (structureboundingbox.isInside(blockposition1)) {
                     this.hasPlacedSpawner = true;
+                    // CraftBukkit start
+                    /*
                     generatoraccessseed.setBlock(blockposition1, Blocks.SPAWNER.defaultBlockState(), 2);
                     TileEntity tileentity = generatoraccessseed.getBlockEntity(blockposition1);
 
@@ -675,6 +677,9 @@
 
                         tileentitymobspawner.setEntityId(EntityTypes.BLAZE, randomsource);
                     }
+                    */
+                    placeCraftSpawner(generatoraccessseed, blockposition1, org.bukkit.entity.EntityType.BLAZE, 2);
+                    // CraftBukkit end
                 }
             }
 
