--- a/net/minecraft/world/level/levelgen/structure/structures/ShipwreckPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/ShipwreckPieces.java
@@ -78,7 +78,12 @@
             ResourceKey<LootTable> resourcekey = (ResourceKey) ShipwreckPieces.MARKERS_TO_LOOT.get(s);
 
             if (resourcekey != null) {
+                // CraftBukkit start - ensure block transformation
+                /*
                 RandomizableContainer.setBlockEntityLootTable(serverlevelaccessor, randomsource, blockpos.below(), resourcekey);
+                */
+                setCraftLootTable(serverlevelaccessor, blockpos.below(), randomsource, resourcekey);
+                // CraftBukkit end
             }
 
         }
