--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -54,7 +54,7 @@
                 public boolean doPlace(int i) {
                     return super.doPlace(i) && i > 5;
                 }
-            }};
+            } }; // CraftBukkit - fix decompile styling
     private static List<StrongholdPieces.PieceWeight> currentPieces;
     private static @Nullable Class<? extends StrongholdPieces.StrongholdPiece> imposedPiece;
     private static int totalWeight;
@@ -1380,6 +1380,8 @@
 
                 if (boundingbox.isInside(blockpos1)) {
                     this.hasPlacedSpawner = true;
+                    // CraftBukkit start
+                    /*
                     worldgenlevel.setBlock(blockpos1, Blocks.SPAWNER.defaultBlockState(), 2);
                     BlockEntity blockentity = worldgenlevel.getBlockEntity(blockpos1);
 
@@ -1388,6 +1390,9 @@
 
                         spawnerblockentity.setEntityId(EntityType.SILVERFISH, randomsource);
                     }
+                    */
+                    placeCraftSpawner(worldgenlevel, blockpos1, org.bukkit.entity.EntityType.SILVERFISH, 2);
+                    // CraftBukkit end
                 }
             }
 
