--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -54,7 +54,7 @@
                 public boolean doPlace(int i) {
                     return super.doPlace(i) && i > 5;
                 }
-            }};
+            } }; // CraftBukkit - fix decompile styling
     private static List<StrongholdPieces.f> currentPieces;
     static Class<? extends StrongholdPieces.p> imposedPiece;
     private static int totalWeight;
@@ -1384,6 +1384,8 @@
 
                 if (structureboundingbox.isInside(blockposition1)) {
                     this.hasPlacedSpawner = true;
+                    // CraftBukkit start
+                    /*
                     generatoraccessseed.setBlock(blockposition1, Blocks.SPAWNER.defaultBlockState(), 2);
                     TileEntity tileentity = generatoraccessseed.getBlockEntity(blockposition1);
 
@@ -1392,6 +1394,9 @@
 
                         tileentitymobspawner.setEntityId(EntityTypes.SILVERFISH, randomsource);
                     }
+                    */
+                    placeCraftSpawner(generatoraccessseed, blockposition1, org.bukkit.entity.EntityType.SILVERFISH, 2);
+                    // CraftBukkit end
                 }
             }
 
