--- a/net/minecraft/world/level/block/BigDripleafBlock.java
+++ b/net/minecraft/world/level/block/BigDripleafBlock.java
@@ -46,6 +46,12 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import net.minecraft.world.entity.player.Player;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
+
 public class BigDripleafBlock extends HorizontalDirectionalBlock implements SimpleWaterloggedBlock, BonemealableBlock {
 
     public static final MapCodec<BigDripleafBlock> CODEC = simpleCodec(BigDripleafBlock::new);
@@ -121,7 +127,7 @@
 
     @Override
     protected void onProjectileHit(Level level, BlockState blockstate, BlockHitResult blockhitresult, Projectile projectile) {
-        this.setTiltAndScheduleTick(blockstate, level, blockhitresult.getBlockPos(), Tilt.FULL, SoundEvents.BIG_DRIPLEAF_TILT_DOWN);
+        this.setTiltAndScheduleTick(blockstate, level, blockhitresult.getBlockPos(), Tilt.FULL, SoundEvents.BIG_DRIPLEAF_TILT_DOWN, projectile); // CraftBukkit
     }
 
     @Override
@@ -180,7 +186,20 @@
     protected void entityInside(BlockState blockstate, Level level, BlockPos blockpos, Entity entity, InsideBlockEffectApplier insideblockeffectapplier, boolean flag) {
         if (!level.isClientSide()) {
             if (blockstate.getValue(BigDripleafBlock.TILT) == Tilt.NONE && canEntityTilt(blockpos, entity) && !level.hasNeighborSignal(blockpos)) {
-                this.setTiltAndScheduleTick(blockstate, level, blockpos, Tilt.UNSTABLE, (SoundEvent) null);
+                // CraftBukkit start - tilt dripleaf
+                org.bukkit.event.Cancellable cancellable;
+                if (entity instanceof Player) {
+                    cancellable = CraftEventFactory.callPlayerInteractEvent((Player) entity, org.bukkit.event.block.Action.PHYSICAL, blockpos, null, null, null);
+                } else {
+                    cancellable = new EntityInteractEvent(entity.getBukkitEntity(), level.getWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+                    level.getCraftServer().getPluginManager().callEvent((EntityInteractEvent) cancellable);
+                }
+
+                if (cancellable.isCancelled()) {
+                    return;
+                }
+                this.setTiltAndScheduleTick(blockstate, level, blockpos, Tilt.UNSTABLE, (SoundEvent) null, entity);
+                // CraftBukkit end
             }
 
         }
@@ -194,9 +213,9 @@
             Tilt tilt = (Tilt) blockstate.getValue(BigDripleafBlock.TILT);
 
             if (tilt == Tilt.UNSTABLE) {
-                this.setTiltAndScheduleTick(blockstate, serverlevel, blockpos, Tilt.PARTIAL, SoundEvents.BIG_DRIPLEAF_TILT_DOWN);
+                this.setTiltAndScheduleTick(blockstate, serverlevel, blockpos, Tilt.PARTIAL, SoundEvents.BIG_DRIPLEAF_TILT_DOWN, null); // CraftBukkit
             } else if (tilt == Tilt.PARTIAL) {
-                this.setTiltAndScheduleTick(blockstate, serverlevel, blockpos, Tilt.FULL, SoundEvents.BIG_DRIPLEAF_TILT_DOWN);
+                this.setTiltAndScheduleTick(blockstate, serverlevel, blockpos, Tilt.FULL, SoundEvents.BIG_DRIPLEAF_TILT_DOWN, null); // CraftBukkit
             } else if (tilt == Tilt.FULL) {
                 resetTilt(blockstate, serverlevel, blockpos);
             }
@@ -222,8 +241,10 @@
         return entity.onGround() && entity.position().y > (double) ((float) blockpos.getY() + 0.6875F);
     }
 
-    private void setTiltAndScheduleTick(BlockState blockstate, Level level, BlockPos blockpos, Tilt tilt, @Nullable SoundEvent soundevent) {
-        setTilt(blockstate, level, blockpos, tilt);
+    // CraftBukkit start
+    private void setTiltAndScheduleTick(BlockState blockstate, Level level, BlockPos blockpos, Tilt tilt, @Nullable SoundEvent soundevent, @Nullable Entity entity) {
+        if (!setTilt(blockstate, level, blockpos, tilt, entity)) return;
+        // CraftBukkit end
         if (soundevent != null) {
             playTiltSound(level, blockpos, soundevent);
         }
@@ -237,14 +258,21 @@
     }
 
     private static void resetTilt(BlockState blockstate, Level level, BlockPos blockpos) {
-        setTilt(blockstate, level, blockpos, Tilt.NONE);
+        setTilt(blockstate, level, blockpos, Tilt.NONE, null); // CraftBukkit
         if (blockstate.getValue(BigDripleafBlock.TILT) != Tilt.NONE) {
             playTiltSound(level, blockpos, SoundEvents.BIG_DRIPLEAF_TILT_UP);
         }
 
     }
 
-    private static void setTilt(BlockState blockstate, Level level, BlockPos blockpos, Tilt tilt) {
+    // CraftBukkit start
+    private static boolean setTilt(BlockState blockstate, Level level, BlockPos blockpos, Tilt tilt, @Nullable Entity entity) {
+        if (entity != null) {
+            if (!CraftEventFactory.callEntityChangeBlockEvent(entity, blockpos, blockstate.setValue(BigDripleafBlock.TILT, tilt))) {
+                return false;
+            }
+        }
+        // CraftBukkit end
         Tilt tilt1 = (Tilt) blockstate.getValue(BigDripleafBlock.TILT);
 
         level.setBlock(blockpos, (BlockState) blockstate.setValue(BigDripleafBlock.TILT, tilt), 2);
@@ -252,6 +280,7 @@
             level.gameEvent((Entity) null, (Holder) GameEvent.BLOCK_CHANGE, blockpos);
         }
 
+        return true; // CraftBukkit
     }
 
     @Override
