--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -412,7 +412,13 @@
                 ItemEntity itementity = (ItemEntity) supplier.get();
 
                 itementity.setDefaultPickUpDelay();
-                level.addFreshEntity(itementity);
+                // CraftBukkit start
+                if (level.captureDrops != null) {
+                    level.captureDrops.add(itementity);
+                } else {
+                    level.addFreshEntity(itementity);
+                }
+                // CraftBukkit end
                 return;
             }
         }
@@ -440,7 +446,7 @@
 
     public void playerDestroy(Level level, Player player, BlockPos blockpos, BlockState blockstate, @Nullable BlockEntity blockentity, ItemStack itemstack) {
         player.awardStat(Stats.BLOCK_MINED.get(this));
-        player.causeFoodExhaustion(0.005F);
+        player.causeFoodExhaustion(0.005F, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.BLOCK_MINED); // CraftBukkit - EntityExhaustionEvent
         dropResources(blockstate, level, blockpos, blockentity, player, itemstack);
     }
 
@@ -547,7 +553,7 @@
     }
 
     protected Function<BlockState, VoxelShape> getShapeForEachState(Function<BlockState, VoxelShape> function) {
-        ImmutableMap immutablemap = (ImmutableMap) this.stateDefinition.getPossibleStates().stream().collect(ImmutableMap.toImmutableMap(Function.identity(), function));
+        ImmutableMap<BlockState, VoxelShape> immutablemap = (ImmutableMap) this.stateDefinition.getPossibleStates().stream().collect(ImmutableMap.toImmutableMap(Function.identity(), function)); // CraftBukkit - decompile error
 
         Objects.requireNonNull(immutablemap);
         return immutablemap::get;
@@ -584,14 +590,22 @@
         return this.builtInRegistryHolder;
     }
 
-    protected void tryDropExperience(ServerLevel serverlevel, BlockPos blockpos, ItemStack itemstack, IntProvider intprovider) {
+    // CraftBukkit start
+    protected int tryDropExperience(ServerLevel serverlevel, BlockPos blockpos, ItemStack itemstack, IntProvider intprovider) {
         int i = EnchantmentHelper.processBlockExperience(serverlevel, itemstack, intprovider.sample(serverlevel.getRandom()));
 
         if (i > 0) {
-            this.popExperience(serverlevel, blockpos, i);
+            // this.popExperience(serverlevel, blockposition, i);
+            return i;
         }
 
+        return 0;
+    }
+
+    public int getExpDrop(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, ItemStack itemstack, boolean flag) {
+        return 0;
     }
+    // CraftBukkit end
 
     private static record ShapePairKey(VoxelShape first, VoxelShape second) {
 
