--- a/net/minecraft/world/level/block/BlockStem.java
+++ b/net/minecraft/world/level/block/BlockStem.java
@@ -27,6 +27,8 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import net.minecraft.world.phys.shapes.VoxelShapeCollision;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class BlockStem extends VegetationBlock implements IBlockFragilePlantElement {
 
     public static final MapCodec<BlockStem> CODEC = RecordCodecBuilder.mapCodec((instance) -> {
@@ -80,7 +82,7 @@
 
                 if (i < 7) {
                     iblockdata = (IBlockData) iblockdata.setValue(BlockStem.AGE, i + 1);
-                    worldserver.setBlock(blockposition, iblockdata, 2);
+                    CraftEventFactory.handleBlockGrowEvent(worldserver, blockposition, iblockdata, 2); // CraftBukkit
                 } else {
                     EnumDirection enumdirection = EnumDirection.EnumDirectionLimit.HORIZONTAL.getRandomDirection(randomsource);
                     BlockPosition blockposition1 = blockposition.relative(enumdirection);
@@ -92,7 +94,11 @@
                         Optional<Block> optional1 = iregistry.getOptional(this.attachedStem);
 
                         if (optional.isPresent() && optional1.isPresent()) {
-                            worldserver.setBlockAndUpdate(blockposition1, ((Block) optional.get()).defaultBlockState());
+                            // CraftBukkit start
+                            if (!CraftEventFactory.handleBlockGrowEvent(worldserver, blockposition1, ((Block) optional.get()).defaultBlockState())) {
+                                return;
+                            }
+                            // CraftBukkit end
                             worldserver.setBlockAndUpdate(blockposition, (IBlockData) ((Block) optional1.get()).defaultBlockState().setValue(BlockFacingHorizontal.FACING, enumdirection));
                         }
                     }
@@ -122,7 +128,7 @@
         int i = Math.min(7, (Integer) iblockdata.getValue(BlockStem.AGE) + MathHelper.nextInt(worldserver.random, 2, 5));
         IBlockData iblockdata1 = (IBlockData) iblockdata.setValue(BlockStem.AGE, i);
 
-        worldserver.setBlock(blockposition, iblockdata1, 2);
+        CraftEventFactory.handleBlockGrowEvent(worldserver, blockposition, iblockdata1, 2); // CraftBukkit
         if (i == 7) {
             iblockdata1.randomTick(worldserver, blockposition, worldserver.random);
         }
