--- a/net/minecraft/world/level/block/CaveVines.java
+++ b/net/minecraft/world/level/block/CaveVines.java
@@ -17,6 +17,15 @@
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import java.util.LinkedList;
+import java.util.List;
+import net.minecraft.world.entity.player.Player;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerHarvestBlockEvent;
+// CraftBukkit end
+
 public interface CaveVines {
 
     VoxelShape SHAPE = Block.column(14.0D, 0.0D, 16.0D);
@@ -27,9 +36,29 @@
             if (level instanceof ServerLevel) {
                 ServerLevel serverlevel = (ServerLevel) level;
 
-                Block.dropFromBlockInteractLootTable(serverlevel, BuiltInLootTables.HARVEST_CAVE_VINE, blockstate, level.getBlockEntity(blockpos), (ItemStack) null, entity, (serverlevel1, itemstack) -> {
-                    Block.popResource(serverlevel1, blockpos, itemstack);
-                });
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockpos, (BlockState) blockstate.setValue(CaveVines.BERRIES, false))) {
+                    return InteractionResult.SUCCESS;
+                }
+
+                if (entity instanceof Player) {
+                    List<ItemStack> dropped = new LinkedList<>();
+                    Block.dropFromBlockInteractLootTable(serverlevel, BuiltInLootTables.HARVEST_CAVE_VINE, blockstate, level.getBlockEntity(blockpos), (ItemStack) null, entity, (serverlevel1, itemstack) -> {
+                        dropped.add(itemstack);
+                    });
+                    PlayerHarvestBlockEvent event = CraftEventFactory.callPlayerHarvestBlockEvent(level, blockpos, (Player) entity, net.minecraft.world.InteractionHand.MAIN_HAND, dropped);
+                    if (event.isCancelled()) {
+                        return InteractionResult.SUCCESS; // We need to return a success either way, because making it PASS or FAIL will result in a bug where cancelling while harvesting w/ block in hand places block
+                    }
+                    for (org.bukkit.inventory.ItemStack itemStack : event.getItemsHarvested()) {
+                        Block.popResource(level, blockpos, CraftItemStack.asNMSCopy(itemStack));
+                    }
+                } else {
+                    Block.dropFromBlockInteractLootTable(serverlevel, BuiltInLootTables.HARVEST_CAVE_VINE, blockstate, level.getBlockEntity(blockpos), (ItemStack) null, entity, (serverlevel1, itemstack) -> {
+                        Block.popResource(serverlevel1, blockpos, itemstack);
+                    });
+                }
+                // CraftBukkit end
                 float f = Mth.randomBetween(serverlevel.random, 0.8F, 1.2F);
 
                 serverlevel.playSound((Entity) null, blockpos, SoundEvents.CAVE_VINES_PICK_BERRIES, SoundSource.BLOCKS, 1.0F, f);
