--- a/net/minecraft/world/level/block/CaveVines.java
+++ b/net/minecraft/world/level/block/CaveVines.java
@@ -17,6 +17,15 @@
 import net.minecraft.world.level.storage.loot.LootTables;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import java.util.LinkedList;
+import java.util.List;
+import net.minecraft.world.entity.player.EntityHuman;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerHarvestBlockEvent;
+// CraftBukkit end
+
 public interface CaveVines {
 
     VoxelShape SHAPE = Block.column(14.0D, 0.0D, 16.0D);
@@ -27,9 +36,29 @@
             if (world instanceof WorldServer) {
                 WorldServer worldserver = (WorldServer) world;
 
-                Block.dropFromBlockInteractLootTable(worldserver, LootTables.HARVEST_CAVE_VINE, iblockdata, world.getBlockEntity(blockposition), (ItemStack) null, entity, (worldserver1, itemstack) -> {
-                    Block.popResource(worldserver1, blockposition, itemstack);
-                });
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition, (IBlockData) iblockdata.setValue(CaveVines.BERRIES, false))) {
+                    return EnumInteractionResult.SUCCESS;
+                }
+
+                if (entity instanceof EntityHuman) {
+                    List<ItemStack> dropped = new LinkedList<>();
+                    Block.dropFromBlockInteractLootTable(worldserver, LootTables.HARVEST_CAVE_VINE, iblockdata, world.getBlockEntity(blockposition), (ItemStack) null, entity, (worldserver1, itemstack) -> {
+                        dropped.add(itemstack);
+                    });
+                    PlayerHarvestBlockEvent event = CraftEventFactory.callPlayerHarvestBlockEvent(world, blockposition, (EntityHuman) entity, net.minecraft.world.EnumHand.MAIN_HAND, dropped);
+                    if (event.isCancelled()) {
+                        return EnumInteractionResult.SUCCESS; // We need to return a success either way, because making it PASS or FAIL will result in a bug where cancelling while harvesting w/ block in hand places block
+                    }
+                    for (org.bukkit.inventory.ItemStack itemStack : event.getItemsHarvested()) {
+                        Block.popResource(world, blockposition, CraftItemStack.asNMSCopy(itemStack));
+                    }
+                } else {
+                    Block.dropFromBlockInteractLootTable(worldserver, LootTables.HARVEST_CAVE_VINE, iblockdata, world.getBlockEntity(blockposition), (ItemStack) null, entity, (worldserver1, itemstack) -> {
+                        Block.popResource(worldserver1, blockposition, itemstack);
+                    });
+                }
+                // CraftBukkit end
                 float f = MathHelper.randomBetween(worldserver.random, 0.8F, 1.2F);
 
                 worldserver.playSound((Entity) null, blockposition, SoundEffects.CAVE_VINES_PICK_BERRIES, SoundCategory.BLOCKS, 1.0F, f);
