--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -19,7 +19,7 @@
 
         if (randomsource.nextFloat() < 0.05688889F) {
             this.getNextState(blockstate, serverlevel, blockpos, randomsource).ifPresent((blockstate1) -> {
-                serverlevel.setBlockAndUpdate(blockpos, blockstate1);
+                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(serverlevel, blockpos, blockstate1); // CraftBukkit
             });
         }
 
