--- a/net/minecraft/world/level/block/ConcretePowderBlock.java
+++ b/net/minecraft/world/level/block/ConcretePowderBlock.java
@@ -16,6 +16,12 @@
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.craftbukkit.block.CraftBlockStates;
+import org.bukkit.event.block.BlockFormEvent;
+// CraftBukkit end
+
 public class ConcretePowderBlock extends FallingBlock {
 
     public static final MapCodec<ConcretePowderBlock> CODEC = RecordCodecBuilder.mapCodec((instance) -> {
@@ -38,7 +44,7 @@
     @Override
     public void onLand(Level level, BlockPos blockpos, BlockState blockstate, BlockState blockstate1, FallingBlockEntity fallingblockentity) {
         if (shouldSolidify(level, blockpos, blockstate1)) {
-            level.setBlock(blockpos, this.concrete.defaultBlockState(), 3);
+            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, blockpos, this.concrete.defaultBlockState(), 3); // CraftBukkit
         }
 
     }
@@ -49,7 +55,24 @@
         BlockPos blockpos = blockplacecontext.getClickedPos();
         BlockState blockstate = blockgetter.getBlockState(blockpos);
 
-        return shouldSolidify(blockgetter, blockpos, blockstate) ? this.concrete.defaultBlockState() : super.getStateForPlacement(blockplacecontext);
+        // CraftBukkit start
+        if (!shouldSolidify(blockgetter, blockpos, blockstate)) {
+            return super.getStateForPlacement(blockplacecontext);
+        }
+
+        // TODO: An event factory call for methods like this
+        CraftBlockState blockState = CraftBlockStates.getBlockState(blockplacecontext.getLevel(), blockpos);
+        blockState.setData(this.concrete.defaultBlockState());
+
+        BlockFormEvent event = new BlockFormEvent(blockState.getBlock(), blockState);
+        blockplacecontext.getLevel().getServer().server.getPluginManager().callEvent(event);
+
+        if (!event.isCancelled()) {
+            return blockState.getHandle();
+        }
+
+        return super.getStateForPlacement(blockplacecontext);
+        // CraftBukkit end
     }
 
     private static boolean shouldSolidify(BlockGetter blockgetter, BlockPos blockpos, BlockState blockstate) {
@@ -82,7 +105,25 @@
 
     @Override
     protected BlockState updateShape(BlockState blockstate, LevelReader levelreader, ScheduledTickAccess scheduledtickaccess, BlockPos blockpos, Direction direction, BlockPos blockpos1, BlockState blockstate1, RandomSource randomsource) {
-        return touchesLiquid(levelreader, blockpos) ? this.concrete.defaultBlockState() : super.updateShape(blockstate, levelreader, scheduledtickaccess, blockpos, direction, blockpos1, blockstate1, randomsource);
+        // CraftBukkit start
+        if (touchesLiquid(levelreader, blockpos)) {
+            // Suppress during worldgen
+            if (!(levelreader instanceof Level level)) {
+                return this.concrete.defaultBlockState();
+            }
+            CraftBlockState blockState = CraftBlockStates.getBlockState(level, blockpos);
+            blockState.setData(this.concrete.defaultBlockState());
+
+            BlockFormEvent event = new BlockFormEvent(blockState.getBlock(), blockState);
+            level.getCraftServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                return blockState.getHandle();
+            }
+        }
+
+        return super.updateShape(blockstate, levelreader, scheduledtickaccess, blockpos, direction, blockpos1, blockstate1, randomsource);
+        // CraftBukkit end
     }
 
     @Override
