--- a/net/minecraft/world/level/block/DaylightDetectorBlock.java
+++ b/net/minecraft/world/level/block/DaylightDetectorBlock.java
@@ -74,6 +74,7 @@
 
         i = Mth.clamp(i, 0, 15);
         if ((Integer) blockstate.getValue(DaylightDetectorBlock.POWER) != i) {
+            i = org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, blockpos, ((Integer) blockstate.getValue(POWER)), i).getNewCurrent(); // CraftBukkit - Call BlockRedstoneEvent
             level.setBlock(blockpos, (BlockState) blockstate.setValue(DaylightDetectorBlock.POWER, i), 3);
         }
 
