--- a/net/minecraft/world/level/block/DirtPathBlock.java
+++ b/net/minecraft/world/level/block/DirtPathBlock.java
@@ -51,6 +51,11 @@
 
     @Override
     protected void tick(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, RandomSource randomsource) {
+        // CraftBukkit start - do not fade if the block is valid here
+        if (blockstate.canSurvive(serverlevel, blockpos)) {
+            return;
+        }
+        // CraftBukkit end
         FarmBlock.turnToDirt((Entity) null, blockstate, serverlevel, blockpos);
     }
 
