--- a/net/minecraft/world/level/block/EyeblossomBlock.java
+++ b/net/minecraft/world/level/block/EyeblossomBlock.java
@@ -88,7 +88,11 @@
         } else {
             EyeblossomBlock.a eyeblossomblock_a = this.type.transform();
 
-            worldserver.setBlock(blockposition, eyeblossomblock_a.state(), 3);
+            // CraftBukkit start - BlockFormEvent
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(worldserver, blockposition, eyeblossomblock_a.state(), 3)) {
+                return false;
+            }
+            // CraftBukkit end
             worldserver.gameEvent(GameEvent.BLOCK_CHANGE, blockposition, GameEvent.a.of(iblockdata));
             eyeblossomblock_a.spawnTransformParticle(worldserver, blockposition, randomsource);
             BlockPosition.betweenClosed(blockposition.offset(-3, -2, -3), blockposition.offset(3, 2, 3)).forEach((blockposition1) -> {
