--- a/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -138,11 +138,14 @@
             boolean flag1 = ((TriState) serverlevel.environmentAttributes().getValue(EnvironmentAttributes.EYEBLOSSOM_OPEN, blockpos)).toBoolean(flag);
 
             if (flag != flag1) {
-                serverlevel.setBlock(blockpos, this.opposite(blockstate), 3);
-                EyeblossomBlock.Type eyeblossomblock_type = EyeblossomBlock.Type.fromBoolean(flag).transform();
+                // CraftBukkit start - BlockFormEvent
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(serverlevel, blockpos, this.opposite(blockstate), 3)) {
+                    EyeblossomBlock.Type eyeblossomblock_type = EyeblossomBlock.Type.fromBoolean(flag).transform();
 
-                eyeblossomblock_type.spawnTransformParticle(serverlevel, blockpos, randomsource);
-                serverlevel.playSound((Entity) null, blockpos, eyeblossomblock_type.longSwitchSound(), SoundSource.BLOCKS, 1.0F, 1.0F);
+                    eyeblossomblock_type.spawnTransformParticle(serverlevel, blockpos, randomsource);
+                    serverlevel.playSound((Entity) null, blockpos, eyeblossomblock_type.longSwitchSound(), SoundSource.BLOCKS, 1.0F, 1.0F);
+                }
+                // CraftBukkit end
             }
         }
 
