--- a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -48,7 +48,7 @@
             BlockPos blockpos1 = blockpos.relative(this.growthDirection);
 
             if (this.canGrowInto(serverlevel.getBlockState(blockpos1))) {
-                serverlevel.setBlockAndUpdate(blockpos1, this.getGrowIntoState(blockstate, serverlevel.random));
+                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(serverlevel, blockpos, blockpos1, this.getGrowIntoState(blockstate, serverlevel.random)); // CraftBukkit
             }
         }
 
