--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -27,6 +27,8 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+import org.bukkit.event.block.LeavesDecayEvent; // CraftBukkit
+
 public abstract class LeavesBlock extends Block implements SimpleWaterloggedBlock {
 
     public static final int DECAY_DISTANCE = 7;
@@ -68,6 +70,14 @@
     @Override
     protected void randomTick(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, RandomSource randomsource) {
         if (this.decaying(blockstate)) {
+            // CraftBukkit start
+            LeavesDecayEvent event = new LeavesDecayEvent(serverlevel.getWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+            serverlevel.getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled() || serverlevel.getBlockState(blockpos).getBlock() != this) {
+                return;
+            }
+            // CraftBukkit end
             dropResources(blockstate, serverlevel, blockpos);
             serverlevel.removeBlock(blockpos, false);
         }
