--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -42,7 +42,7 @@
 public class LiquidBlock extends Block implements BucketPickup {
 
     private static final Codec<FlowingFluid> FLOWING_FLUID = BuiltInRegistries.FLUID.byNameCodec().comapFlatMap((fluid) -> {
-        DataResult dataresult;
+        DataResult<FlowingFluid> dataresult; // CraftBukkit - decompile error
 
         if (fluid instanceof FlowingFluid flowingfluid) {
             dataresult = DataResult.success(flowingfluid);
@@ -175,14 +175,20 @@
                 if (level.getFluidState(blockpos1).is(FluidTags.WATER)) {
                     Block block = level.getFluidState(blockpos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
 
-                    level.setBlockAndUpdate(blockpos, block.defaultBlockState());
-                    this.fizz(level, blockpos);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, blockpos, block.defaultBlockState())) {
+                        this.fizz(level, blockpos);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
 
                 if (flag && level.getBlockState(blockpos1).is(Blocks.BLUE_ICE)) {
-                    level.setBlockAndUpdate(blockpos, Blocks.BASALT.defaultBlockState());
-                    this.fizz(level, blockpos);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, blockpos, Blocks.BASALT.defaultBlockState())) {
+                        this.fizz(level, blockpos);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
             }
