--- a/net/minecraft/world/level/block/MagmaBlock.java
+++ b/net/minecraft/world/level/block/MagmaBlock.java
@@ -30,7 +30,7 @@
     @Override
     public void stepOn(Level level, BlockPos blockpos, BlockState blockstate, Entity entity) {
         if (!entity.isSteppingCarefully() && entity instanceof LivingEntity) {
-            entity.hurt(level.damageSources().hotFloor(), 1.0F);
+            entity.hurt(level.damageSources().hotFloor().directBlock(level, blockpos), 1.0F); // CraftBukkit
         }
 
         super.stepOn(level, blockpos, blockstate, entity);
