--- a/net/minecraft/world/level/block/NetherWartBlock.java
+++ b/net/minecraft/world/level/block/NetherWartBlock.java
@@ -56,7 +56,7 @@
 
         if (i < 3 && randomsource.nextInt(10) == 0) {
             blockstate = (BlockState) blockstate.setValue(NetherWartBlock.AGE, i + 1);
-            serverlevel.setBlock(blockpos, blockstate, 2);
+            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(serverlevel, blockpos, blockstate, 2); // CraftBukkit
         }
 
     }
