--- a/net/minecraft/world/level/block/NoteBlock.java
+++ b/net/minecraft/world/level/block/NoteBlock.java
@@ -85,6 +85,7 @@
         if (flag1 != (Boolean) blockstate.getValue(NoteBlock.POWERED)) {
             if (flag1) {
                 this.playNote((Entity) null, blockstate, level, blockpos);
+                blockstate = level.getBlockState(blockpos); // CraftBukkit - SPIGOT-5617: update in case changed in event
             }
 
             level.setBlock(blockpos, (BlockState) blockstate.setValue(NoteBlock.POWERED, flag1), 3);
@@ -94,6 +95,12 @@
 
     private void playNote(@Nullable Entity entity, BlockState blockstate, Level level, BlockPos blockpos) {
         if (((NoteBlockInstrument) blockstate.getValue(NoteBlock.INSTRUMENT)).worksAboveNoteBlock() || level.getBlockState(blockpos.above()).isAir()) {
+            // CraftBukkit start
+            org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(level, blockpos, blockstate.getValue(NoteBlock.INSTRUMENT), blockstate.getValue(NoteBlock.NOTE));
+            if (event.isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
             level.blockEvent(blockpos, this, 0, 0);
             level.gameEvent(entity, (Holder) GameEvent.NOTE_BLOCK_PLAY, blockpos);
         }
