--- a/net/minecraft/world/level/block/ObserverBlock.java
+++ b/net/minecraft/world/level/block/ObserverBlock.java
@@ -18,6 +18,8 @@
 import net.minecraft.world.level.redstone.ExperimentalRedstoneUtils;
 import net.minecraft.world.level.redstone.Orientation;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class ObserverBlock extends DirectionalBlock {
 
     public static final MapCodec<ObserverBlock> CODEC = simpleCodec(ObserverBlock::new);
@@ -51,8 +53,18 @@
     @Override
     protected void tick(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, RandomSource randomsource) {
         if ((Boolean) blockstate.getValue(ObserverBlock.POWERED)) {
+            // CraftBukkit start
+            if (CraftEventFactory.callRedstoneChange(serverlevel, blockpos, 15, 0).getNewCurrent() != 0) {
+                return;
+            }
+            // CraftBukkit end
             serverlevel.setBlock(blockpos, (BlockState) blockstate.setValue(ObserverBlock.POWERED, false), 2);
         } else {
+            // CraftBukkit start
+            if (CraftEventFactory.callRedstoneChange(serverlevel, blockpos, 0, 15).getNewCurrent() != 15) {
+                return;
+            }
+            // CraftBukkit end
             serverlevel.setBlock(blockpos, (BlockState) blockstate.setValue(ObserverBlock.POWERED, true), 2);
             serverlevel.scheduleTick(blockpos, (Block) this, 2);
         }
