--- a/net/minecraft/world/level/block/PowderSnowBlock.java
+++ b/net/minecraft/world/level/block/PowderSnowBlock.java
@@ -77,7 +77,12 @@
 
         insideblockeffectapplier.runBefore(InsideBlockEffectType.EXTINGUISH, (entity1) -> {
             if (world instanceof WorldServer worldserver) {
-                if (entity1.isOnFire() && (worldserver.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) || entity1 instanceof EntityHuman) && entity1.mayInteract(worldserver, blockposition1)) {
+                // CraftBukkit start
+                if (entity1.isOnFire() && entity1.mayInteract(worldserver, blockposition1)) {
+                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity1, blockposition1, Blocks.AIR.defaultBlockState(), !(worldserver.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) || entity1 instanceof EntityHuman))) {
+                        return;
+                    }
+                    // CraftBukkit end
                     world.destroyBlock(blockposition1, false);
                 }
             }
