--- a/net/minecraft/world/level/block/RedstoneLampBlock.java
+++ b/net/minecraft/world/level/block/RedstoneLampBlock.java
@@ -13,6 +13,8 @@
 import net.minecraft.world.level.redstone.Orientation;
 import org.jspecify.annotations.Nullable;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class RedstoneLampBlock extends Block {
 
     public static final MapCodec<RedstoneLampBlock> CODEC = simpleCodec(RedstoneLampBlock::new);
@@ -42,6 +44,11 @@
                 if (flag1) {
                     level.scheduleTick(blockpos, (Block) this, 4);
                 } else {
+                    // CraftBukkit start
+                    if (CraftEventFactory.callRedstoneChange(level, blockpos, 0, 15).getNewCurrent() != 15) {
+                        return;
+                    }
+                    // CraftBukkit end
                     level.setBlock(blockpos, (BlockState) blockstate.cycle(RedstoneLampBlock.LIT), 2);
                 }
             }
@@ -52,6 +59,11 @@
     @Override
     protected void tick(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, RandomSource randomsource) {
         if ((Boolean) blockstate.getValue(RedstoneLampBlock.LIT) && !serverlevel.hasNeighborSignal(blockpos)) {
+            // CraftBukkit start
+            if (CraftEventFactory.callRedstoneChange(serverlevel, blockpos, 15, 0).getNewCurrent() != 0) {
+                return;
+            }
+            // CraftBukkit end
             serverlevel.setBlock(blockpos, (BlockState) blockstate.cycle(RedstoneLampBlock.LIT), 2);
         }
 
