--- a/net/minecraft/world/level/block/RedStoneOreBlock.java
+++ b/net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -21,6 +21,11 @@
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
 import net.minecraft.world.phys.BlockHitResult;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
+
 public class RedStoneOreBlock extends Block {
 
     public static final MapCodec<RedStoneOreBlock> CODEC = simpleCodec(RedStoneOreBlock::new);
@@ -38,14 +43,27 @@
 
     @Override
     protected void attack(BlockState blockstate, Level level, BlockPos blockpos, Player player) {
-        interact(blockstate, level, blockpos);
+        interact(blockstate, level, blockpos, player); // CraftBukkit - add entityhuman
         super.attack(blockstate, level, blockpos, player);
     }
 
     @Override
     public void stepOn(Level level, BlockPos blockpos, BlockState blockstate, Entity entity) {
         if (!entity.isSteppingCarefully()) {
-            interact(blockstate, level, blockpos);
+            // CraftBukkit start
+            if (entity instanceof Player) {
+                org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent((Player) entity, org.bukkit.event.block.Action.PHYSICAL, blockpos, null, null, null);
+                if (!event.isCancelled()) {
+                    interact(level.getBlockState(blockpos), level, blockpos, entity); // add entity
+                }
+            } else {
+                EntityInteractEvent event = new EntityInteractEvent(entity.getBukkitEntity(), level.getWorld().getBlockAt(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+                level.getCraftServer().getPluginManager().callEvent(event);
+                if (!event.isCancelled()) {
+                    interact(level.getBlockState(blockpos), level, blockpos, entity); // add entity
+                }
+            }
+            // CraftBukkit end
         }
 
         super.stepOn(level, blockpos, blockstate, entity);
@@ -56,15 +74,20 @@
         if (level.isClientSide()) {
             spawnParticles(level, blockpos);
         } else {
-            interact(blockstate, level, blockpos);
+            interact(blockstate, level, blockpos, player); // CraftBukkit - add entityhuman
         }
 
         return (InteractionResult) (itemstack.getItem() instanceof BlockItem && (new BlockPlaceContext(player, interactionhand, itemstack, blockhitresult)).canPlace() ? InteractionResult.PASS : InteractionResult.SUCCESS);
     }
 
-    private static void interact(BlockState blockstate, Level level, BlockPos blockpos) {
+    private static void interact(BlockState blockstate, Level level, BlockPos blockpos, Entity entity) { // CraftBukkit - add Entity
         spawnParticles(level, blockpos);
         if (!(Boolean) blockstate.getValue(RedStoneOreBlock.LIT)) {
+            // CraftBukkit start
+            if (!CraftEventFactory.callEntityChangeBlockEvent(entity, blockpos, blockstate.setValue(RedStoneOreBlock.LIT, true))) {
+                return;
+            }
+            // CraftBukkit end
             level.setBlock(blockpos, (BlockState) blockstate.setValue(RedStoneOreBlock.LIT, true), 3);
         }
 
@@ -78,6 +101,11 @@
     @Override
     protected void randomTick(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, RandomSource randomsource) {
         if ((Boolean) blockstate.getValue(RedStoneOreBlock.LIT)) {
+            // CraftBukkit start
+            if (CraftEventFactory.callBlockFadeEvent(serverlevel, blockpos, blockstate.setValue(RedStoneOreBlock.LIT, false)).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
             serverlevel.setBlock(blockpos, (BlockState) blockstate.setValue(RedStoneOreBlock.LIT, false), 3);
         }
 
@@ -86,10 +114,17 @@
     @Override
     protected void spawnAfterBreak(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, ItemStack itemstack, boolean flag) {
         super.spawnAfterBreak(blockstate, serverlevel, blockpos, itemstack, flag);
+        // CraftBukkit start - Delegated to getExpDrop
+    }
+
+    @Override
+    public int getExpDrop(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, ItemStack itemstack, boolean flag) {
         if (flag) {
-            this.tryDropExperience(serverlevel, blockpos, itemstack, UniformInt.of(1, 5));
+            return this.tryDropExperience(serverlevel, blockpos, itemstack, UniformInt.of(1, 5));
         }
 
+        return 0;
+        // CraftBukkit end
     }
 
     @Override
