--- a/net/minecraft/world/level/block/ScaffoldingBlock.java
+++ b/net/minecraft/world/level/block/ScaffoldingBlock.java
@@ -102,7 +102,7 @@
         int i = getDistance(serverlevel, blockpos);
         BlockState blockstate1 = (BlockState) ((BlockState) blockstate.setValue(ScaffoldingBlock.DISTANCE, i)).setValue(ScaffoldingBlock.BOTTOM, this.isBottom(serverlevel, blockpos, i));
 
-        if ((Integer) blockstate1.getValue(ScaffoldingBlock.DISTANCE) == 7) {
+        if ((Integer) blockstate1.getValue(ScaffoldingBlock.DISTANCE) == 7 && !org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(serverlevel, blockpos, Blocks.AIR.defaultBlockState()).isCancelled()) { // CraftBukkit - BlockFadeEvent
             if ((Integer) blockstate.getValue(ScaffoldingBlock.DISTANCE) == 7) {
                 FallingBlockEntity.fall(serverlevel, blockpos, blockstate1);
             } else {
