--- a/net/minecraft/world/level/block/SculkShriekerBlock.java
+++ b/net/minecraft/world/level/block/SculkShriekerBlock.java
@@ -63,6 +63,7 @@
             ServerPlayer serverplayer = SculkShriekerBlockEntity.tryGetPlayer(entity);
 
             if (serverplayer != null) {
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent(serverplayer, org.bukkit.event.block.Action.PHYSICAL, blockpos, null, null, null).isCancelled()) return; // CraftBukkit
                 serverlevel.getBlockEntity(blockpos, BlockEntityType.SCULK_SHRIEKER).ifPresent((sculkshriekerblockentity) -> {
                     sculkshriekerblockentity.tryShriek(serverlevel, serverplayer);
                 });
@@ -125,10 +126,17 @@
     @Override
     protected void spawnAfterBreak(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, ItemStack itemstack, boolean flag) {
         super.spawnAfterBreak(blockstate, serverlevel, blockpos, itemstack, flag);
+        // CraftBukkit start - Delegate to getExpDrop
+    }
+
+    @Override
+    public int getExpDrop(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, ItemStack itemstack, boolean flag) {
         if (flag) {
-            this.tryDropExperience(serverlevel, blockpos, itemstack, ConstantInt.of(5));
+            return this.tryDropExperience(serverlevel, blockpos, itemstack, ConstantInt.of(5));
         }
 
+        return 0;
+        // CraftBukkit end
     }
 
     @Override
