--- a/net/minecraft/world/level/block/StemBlock.java
+++ b/net/minecraft/world/level/block/StemBlock.java
@@ -27,6 +27,8 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class StemBlock extends VegetationBlock implements BonemealableBlock {
 
     public static final MapCodec<StemBlock> CODEC = RecordCodecBuilder.mapCodec((instance) -> {
@@ -80,7 +82,7 @@
 
                 if (i < 7) {
                     blockstate = (BlockState) blockstate.setValue(StemBlock.AGE, i + 1);
-                    serverlevel.setBlock(blockpos, blockstate, 2);
+                    CraftEventFactory.handleBlockGrowEvent(serverlevel, blockpos, blockstate, 2); // CraftBukkit
                 } else {
                     Direction direction = Direction.Plane.HORIZONTAL.getRandomDirection(randomsource);
                     BlockPos blockpos1 = blockpos.relative(direction);
@@ -92,7 +94,11 @@
                         Optional<Block> optional1 = registry.getOptional(this.attachedStem);
 
                         if (optional.isPresent() && optional1.isPresent()) {
-                            serverlevel.setBlockAndUpdate(blockpos1, ((Block) optional.get()).defaultBlockState());
+                            // CraftBukkit start
+                            if (!CraftEventFactory.handleBlockGrowEvent(serverlevel, blockpos1, ((Block) optional.get()).defaultBlockState())) {
+                                return;
+                            }
+                            // CraftBukkit end
                             serverlevel.setBlockAndUpdate(blockpos, (BlockState) ((Block) optional1.get()).defaultBlockState().setValue(HorizontalDirectionalBlock.FACING, direction));
                         }
                     }
@@ -122,7 +128,7 @@
         int i = Math.min(7, (Integer) blockstate.getValue(StemBlock.AGE) + Mth.nextInt(serverlevel.random, 2, 5));
         BlockState blockstate1 = (BlockState) blockstate.setValue(StemBlock.AGE, i);
 
-        serverlevel.setBlock(blockpos, blockstate1, 2);
+        CraftEventFactory.handleBlockGrowEvent(serverlevel, blockpos, blockstate1, 2); // CraftBukkit
         if (i == 7) {
             blockstate1.randomTick(serverlevel, blockpos, serverlevel.random);
         }
