--- a/net/minecraft/world/level/block/VineBlock.java
+++ b/net/minecraft/world/level/block/VineBlock.java
@@ -23,6 +23,8 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class VineBlock extends Block {
 
     public static final MapCodec<VineBlock> CODEC = simpleCodec(VineBlock::new);
@@ -178,30 +180,34 @@
                             BlockPos blockpos3 = blockpos2.relative(direction1);
                             BlockPos blockpos4 = blockpos2.relative(direction2);
 
+                            // CraftBukkit start - Call BlockSpreadEvent
+                            BlockPos source = blockpos;
+
                             if (flag && isAcceptableNeighbour(serverlevel, blockpos3, direction1)) {
-                                serverlevel.setBlock(blockpos2, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction1), true), 2);
+                                CraftEventFactory.handleBlockSpreadEvent(serverlevel, source, blockpos2, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction1), true), 2);
                             } else if (flag1 && isAcceptableNeighbour(serverlevel, blockpos4, direction2)) {
-                                serverlevel.setBlock(blockpos2, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction2), true), 2);
+                                CraftEventFactory.handleBlockSpreadEvent(serverlevel, source, blockpos2, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction2), true), 2);
                             } else {
                                 Direction direction3 = direction.getOpposite();
 
                                 if (flag && serverlevel.isEmptyBlock(blockpos3) && isAcceptableNeighbour(serverlevel, blockpos.relative(direction1), direction3)) {
-                                    serverlevel.setBlock(blockpos3, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction3), true), 2);
+                                    CraftEventFactory.handleBlockSpreadEvent(serverlevel, source, blockpos3, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction3), true), 2);
                                 } else if (flag1 && serverlevel.isEmptyBlock(blockpos4) && isAcceptableNeighbour(serverlevel, blockpos.relative(direction2), direction3)) {
-                                    serverlevel.setBlock(blockpos4, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction3), true), 2);
+                                    CraftEventFactory.handleBlockSpreadEvent(serverlevel, source, blockpos4, (BlockState) this.defaultBlockState().setValue(getPropertyForFace(direction3), true), 2);
                                 } else if ((double) randomsource.nextFloat() < 0.05D && isAcceptableNeighbour(serverlevel, blockpos2.above(), Direction.UP)) {
-                                    serverlevel.setBlock(blockpos2, (BlockState) this.defaultBlockState().setValue(VineBlock.UP, true), 2);
+                                    CraftEventFactory.handleBlockSpreadEvent(serverlevel, source, blockpos2, (BlockState) this.defaultBlockState().setValue(VineBlock.UP, true), 2);
                                 }
+                                // CraftBukkit end
                             }
                         } else if (isAcceptableNeighbour(serverlevel, blockpos2, direction)) {
-                            serverlevel.setBlock(blockpos, (BlockState) blockstate.setValue(getPropertyForFace(direction), true), 2);
+                            CraftEventFactory.handleBlockGrowEvent(serverlevel, blockpos, (BlockState) blockstate.setValue(getPropertyForFace(direction), true), 2); // CraftBukkit
                         }
 
                     }
                 } else {
                     if (direction == Direction.UP && blockpos.getY() < serverlevel.getMaxY()) {
                         if (this.canSupportAtFace(serverlevel, blockpos, direction)) {
-                            serverlevel.setBlock(blockpos, (BlockState) blockstate.setValue(VineBlock.UP, true), 2);
+                            CraftEventFactory.handleBlockGrowEvent(serverlevel, blockpos, (BlockState) blockstate.setValue(VineBlock.UP, true), 2); // CraftBukkit
                             return;
                         }
 
@@ -219,7 +225,7 @@
                             }
 
                             if (this.hasHorizontalConnection(blockstate2)) {
-                                serverlevel.setBlock(blockpos1, blockstate2, 2);
+                                CraftEventFactory.handleBlockSpreadEvent(serverlevel, blockpos, blockpos1, blockstate2, 2); // CraftBukkit
                             }
 
                             return;
@@ -235,7 +241,7 @@
                             BlockState blockstate5 = this.copyRandomFaces(blockstate, blockstate4, randomsource);
 
                             if (blockstate4 != blockstate5 && this.hasHorizontalConnection(blockstate5)) {
-                                serverlevel.setBlock(blockpos5, blockstate5, 2);
+                                CraftEventFactory.handleBlockSpreadEvent(serverlevel, blockpos, blockpos5, blockstate5, 2); // CraftBukkit
                             }
                         }
                     }
