--- /dev/null
+++ b/net/minecraft/client/model/monster/zombie/AbstractZombieModel.java
@@ -1,0 +_,20 @@
+package net.minecraft.client.model.monster.zombie;
+
+import net.minecraft.client.model.AnimationUtils;
+import net.minecraft.client.model.HumanoidModel;
+import net.minecraft.client.model.geom.ModelPart;
+import net.minecraft.client.renderer.entity.state.ZombieRenderState;
+import net.neoforged.api.distmarker.Dist;
+import net.neoforged.api.distmarker.OnlyIn;
+
+@OnlyIn(Dist.CLIENT)
+public abstract class AbstractZombieModel<S extends ZombieRenderState> extends HumanoidModel<S> {
+    protected AbstractZombieModel(ModelPart root) {
+        super(root);
+    }
+
+    public void setupAnim(S state) {
+        super.setupAnim(state);
+        AnimationUtils.animateZombieArms(this.leftArm, this.rightArm, state.isAggressive, state);
+    }
+}
