--- a/net/minecraft/client/model/monster/zombie/DrownedModel.java
+++ b/net/minecraft/client/model/monster/zombie/DrownedModel.java
@@ -33,20 +_,39 @@
     }
 
     @Override
-    protected void setupAttackAnimation(ZombieRenderState state) {
-        super.setupAttackAnimation(state);
-        AnimationUtils.animateZombieArms(this.leftArm, this.rightArm, state.isAggressive, state);
-    }
-
-    protected void setupSwimAnimation(ZombieRenderState state, float animationPos, float swimAmount) {
-        this.rightArm.xRot = Mth.rotLerpRad(swimAmount, this.rightArm.xRot, (float) (-Math.PI * 4.0 / 5.0))
-            + swimAmount * 0.35F * Mth.sin(0.1F * state.ageInTicks);
-        this.leftArm.xRot = Mth.rotLerpRad(swimAmount, this.leftArm.xRot, (float) (-Math.PI * 4.0 / 5.0))
-            - swimAmount * 0.35F * Mth.sin(0.1F * state.ageInTicks);
-        this.rightArm.zRot = Mth.rotLerpRad(swimAmount, this.rightArm.zRot, -0.15F);
-        this.leftArm.zRot = Mth.rotLerpRad(swimAmount, this.leftArm.zRot, 0.15F);
-        this.leftLeg.xRot = this.leftLeg.xRot - swimAmount * 0.55F * Mth.sin(0.1F * state.ageInTicks);
-        this.rightLeg.xRot = this.rightLeg.xRot + swimAmount * 0.55F * Mth.sin(0.1F * state.ageInTicks);
-        this.head.xRot = 0.0F;
+    public void setupAnim(ZombieRenderState state) {
+        super.setupAnim(state);
+        if (state.leftArmPose == HumanoidModel.ArmPose.THROW_TRIDENT) {
+            this.leftArm.xRot = this.leftArm.xRot * 0.5F - (float) Math.PI;
+            this.leftArm.yRot = 0.0F;
+        }
+
+        if (state.rightArmPose == HumanoidModel.ArmPose.THROW_TRIDENT) {
+            this.rightArm.xRot = this.rightArm.xRot * 0.5F - (float) Math.PI;
+            this.rightArm.yRot = 0.0F;
+        }
+
+        float swimAmount = state.swimAmount;
+        if (swimAmount > 0.0F) {
+            this.rightArm.xRot = Mth.rotLerpRad(swimAmount, this.rightArm.xRot, (float) (-Math.PI * 4.0 / 5.0))
+                + swimAmount * 0.35F * Mth.sin(0.1F * state.ageInTicks);
+            this.leftArm.xRot = Mth.rotLerpRad(swimAmount, this.leftArm.xRot, (float) (-Math.PI * 4.0 / 5.0))
+                - swimAmount * 0.35F * Mth.sin(0.1F * state.ageInTicks);
+            this.rightArm.zRot = Mth.rotLerpRad(swimAmount, this.rightArm.zRot, -0.15F);
+            this.leftArm.zRot = Mth.rotLerpRad(swimAmount, this.leftArm.zRot, 0.15F);
+            this.leftLeg.xRot = this.leftLeg.xRot - swimAmount * 0.55F * Mth.sin(0.1F * state.ageInTicks);
+            this.rightLeg.xRot = this.rightLeg.xRot + swimAmount * 0.55F * Mth.sin(0.1F * state.ageInTicks);
+            this.head.xRot = 0.0F;
+        } else if (state.leftArmPose == HumanoidModel.ArmPose.THROW_TRIDENT) {
+            this.leftArm.xRot += 1.2F;
+        } else if (state.rightArmPose == HumanoidModel.ArmPose.THROW_TRIDENT) {
+            this.rightArm.xRot += 1.2F;
+        }
+
+        if (state.attackTime > 0.0F) {
+            AnimationUtils.animateZombieArms(this.leftArm, this.rightArm, state.isAggressive, state);
+            this.rightArm.xRot -= 1.2F;
+            this.leftArm.xRot -= 1.2F;
+        }
     }
 }
