--- a/net/minecraft/client/model/monster/zombie/ZombieVillagerModel.java
+++ b/net/minecraft/client/model/monster/zombie/ZombieVillagerModel.java
@@ -96,8 +_,8 @@
         return mesh;
     }
 
-    protected void setupAttackAnimation(S state) {
-        super.setupAttackAnimation(state);
+    public void setupAnim(S state) {
+        super.setupAnim(state);
         AnimationUtils.animateZombieArms(this.leftArm, this.rightArm, state.isAggressive, state);
     }
 
