--- a/net/minecraft/world/item/component/AttackRange.java
+++ b/net/minecraft/world/item/component/AttackRange.java
@@ -25,6 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 
 public record AttackRange(float minReach, float maxReach, float minCreativeReach, float maxCreativeReach, float hitboxMargin, float mobFactor) {
+    public static final AttackRange CODEC_DEFAULT = new AttackRange(0.0F, 3.0F, 0.0F, 5.0F, 0.3F, 1.0F); // Paper - add back defaults instance
     public static final Codec<AttackRange> CODEC = RecordCodecBuilder.create(
         i -> i.group(
                 ExtraCodecs.floatRange(0.0F, 64.0F).optionalFieldOf("min_reach", 0.0F).forGetter(AttackRange::minReach),
