--- a/net/minecraft/world/item/component/BundleContents.java
+++ b/net/minecraft/world/item/component/BundleContents.java
@@ -73,7 +_,8 @@
     }
 
     public static boolean canItemBeInBundle(ItemStack itemsToAdd) {
-        return !itemsToAdd.isEmpty() && itemsToAdd.getItem().canFitInsideContainerItems();
+        // Neo: stack-aware placeability check
+        return !itemsToAdd.isEmpty() && itemsToAdd.canFitInsideContainerItems();
     }
 
     public int getNumberOfItemsToShow() {
