--- a/net/minecraft/world/item/component/BundleContents.java
+++ b/net/minecraft/world/item/component/BundleContents.java
@@ -29,9 +_,10 @@
 
 public final class BundleContents implements ContainerComponent<BundleContents>, TooltipComponent {
     public static final BundleContents EMPTY = new BundleContents(List.of());
-    public static final Codec<BundleContents> CODEC = ItemStackTemplate.CODEC.listOf().xmap(BundleContents::new, contents -> contents.items);
+    public static final Codec<BundleContents> CODEC = ItemStackTemplate.CODEC.sizeLimitedListOf(256).xmap(BundleContents::new, contents -> contents.items); // Paper - limit size
     public static final StreamCodec<RegistryFriendlyByteBuf, BundleContents> STREAM_CODEC = ItemStackTemplate.STREAM_CODEC
-        .apply(ByteBufCodecs.list())
+        .apply(ByteBufCodecs.list(256)) // Paper - limit size
+        .apply(ByteBufCodecs::increaseDepth) // Paper - Track codec depth
         .map(BundleContents::new, contents -> contents.items);
     private static final Fraction BUNDLE_IN_BUNDLE_WEIGHT = Fraction.getFraction(1, 16);
     private static final int NO_STACK_INDEX = -1;
@@ -42,6 +_,11 @@
     private final Supplier<DataResult<Fraction>> weight;
 
     private BundleContents(List<ItemStackTemplate> items, int selectedItem) {
+        // Paper start - limit size
+        if (items.size() > 256) {
+            throw new IllegalArgumentException("Too many items");
+        }
+        // Paper end - limit size
         this.items = items;
         this.selectedItem = selectedItem;
         this.weight = Suppliers.memoize(() -> computeContentWeight(this.items));
@@ -81,8 +_,16 @@
     }
 
     public static boolean canItemBeInBundle(ItemStack itemToAdd) {
-        return !itemToAdd.isEmpty() && itemToAdd.getItem().canFitInsideContainerItems();
-    }
+        // Neo: stack-aware placeability check
+        return !itemToAdd.isEmpty() && itemToAdd.canFitInsideContainerItems();
+    }
+
+    // Paper start - correct bundle inventory action
+    public int getMaxAmountToAdd(final ItemStack stack) {
+        DataResult<Fraction> weight = this.weight();
+        return weight.isError() ? 0 : Mutable.getMaxAmountToAdd(stack, weight.getOrThrow());
+    }
+    // Paper end - correct bundle inventory action
 
     public int getNumberOfItemsToShow() {
         int numberOfItemStacks = this.size();
@@ -162,7 +_,7 @@
         return "BundleContents" + this.items;
     }
 
-    public static class Mutable extends GrowableMutableContainer<BundleContents> {
+    public static class Mutable extends SimpleMutableContainer<BundleContents> {
         private Fraction weight;
         private int selectedItem;
         private boolean needsFlattening;
@@ -210,6 +_,14 @@
             return Math.max(remainingWeight.divideBy(itemWeight).intValue(), 0);
         }
 
+        // Paper start - correct bundle inventory action
+        // Static overload to easily compute this value without the need for an instance of mutable.
+        static int getMaxAmountToAdd(final ItemStack stack, final Fraction weight) {
+            Fraction remainingWeight = Fraction.ONE.subtract(weight);
+            return Math.max(remainingWeight.divideBy(BundleContents.getWeight(stack).getOrThrow()).intValue(), 0);
+        }
+        // Paper end - correct bundle inventory action
+
         public int tryInsert(ItemStack itemsToAdd) {
             if (!BundleContents.canItemBeInBundle(itemsToAdd)) {
                 return 0;
@@ -296,6 +_,21 @@
         }
 
         @Override
+        protected int insertNewSlots(ItemProvider newItems, SlotSelector slotSelector) {
+            int successCount;
+            for (successCount = 0; newItems.hasNext() && this.canInsertNewSlots() && slotSelector.trySelectSlot(ItemStack.EMPTY); successCount++) {
+                boolean success = newItems.findNextNonEmpty() && this.addSlotWithItem(newItems.peek());
+                if (!success) {
+                    break;
+                }
+
+                newItems.next();
+            }
+
+            return successCount;
+        }
+
+        @Override
         protected boolean setItem(int slot, ItemStack itemStack) {
             ItemStack currentItem = this.items.get(slot);
             Fraction adjustedWeight = currentItem.isEmpty() ? this.weight : this.weight.subtract(getStackedWeight(currentItem));
@@ -310,12 +_,12 @@
         }
 
         @Override
-        protected boolean addSlotWithItem(ItemProvider newItems) {
-            if (!newItems.findNextNonEmpty()) {
+        protected boolean addSlotWithItem(ItemStack itemStack) {
+            if (itemStack.isEmpty()) {
                 return false;
             } else {
-                Fraction newWeight = getWeightWithAddedItems(this.weight, newItems.peek());
-                if (newWeight != null && super.addSlotWithItem(newItems)) {
+                Fraction newWeight = getWeightWithAddedItems(this.weight, itemStack);
+                if (newWeight != null && super.addSlotWithItem(itemStack)) {
                     this.weight = newWeight;
                     this.needsFlattening = true;
                     return true;
