--- a/net/minecraft/world/item/component/BundleContents.java
+++ b/net/minecraft/world/item/component/BundleContents.java
@@ -7,31 +_,28 @@
 import com.mojang.serialization.DataResult;
 import java.util.ArrayList;
 import java.util.List;
-import java.util.function.Consumer;
 import java.util.function.Supplier;
 import java.util.stream.Stream;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
-import net.minecraft.world.entity.SlotAccess;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.inventory.Slot;
 import net.minecraft.world.inventory.tooltip.TooltipComponent;
 import net.minecraft.world.item.ItemInstance;
-import net.minecraft.world.item.ItemProvider;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.ItemStackTemplate;
-import net.minecraft.world.item.slot.SlotSelector;
 import net.minecraft.world.level.block.entity.BeehiveBlockEntity;
 import org.apache.commons.lang3.math.Fraction;
 import org.jspecify.annotations.Nullable;
 
-public final class BundleContents implements ContainerComponent<BundleContents>, TooltipComponent {
+public final class BundleContents implements TooltipComponent {
     public static final BundleContents EMPTY = new BundleContents(List.of());
-    public static final Codec<BundleContents> CODEC = ItemStackTemplate.CODEC.listOf().xmap(BundleContents::new, contents -> contents.items);
+    public static final Codec<BundleContents> CODEC = ItemStackTemplate.CODEC.sizeLimitedListOf(256).xmap(BundleContents::new, contents -> contents.items); // Paper - limit size
     public static final StreamCodec<RegistryFriendlyByteBuf, BundleContents> STREAM_CODEC = ItemStackTemplate.STREAM_CODEC
-        .apply(ByteBufCodecs.list())
+        .apply(ByteBufCodecs.list(256)) // Paper - limit size
+        .apply(ByteBufCodecs::increaseDepth) // Paper - Track codec depth
         .map(BundleContents::new, contents -> contents.items);
     private static final Fraction BUNDLE_IN_BUNDLE_WEIGHT = Fraction.getFraction(1, 16);
     private static final int NO_STACK_INDEX = -1;
@@ -42,6 +_,11 @@
     private final Supplier<DataResult<Fraction>> weight;
 
     private BundleContents(List<ItemStackTemplate> items, int selectedItem) {
+        // Paper start - limit size
+        if (items.size() > 256) {
+            throw new IllegalArgumentException("Too many items");
+        }
+        // Paper end - limit size
         this.items = items;
         this.selectedItem = selectedItem;
         this.weight = Suppliers.memoize(() -> computeContentWeight(this.items));
@@ -81,8 +_,16 @@
     }
 
     public static boolean canItemBeInBundle(ItemStack itemToAdd) {
-        return !itemToAdd.isEmpty() && itemToAdd.getItem().canFitInsideContainerItems();
-    }
+        // Neo: stack-aware placeability check
+        return !itemToAdd.isEmpty() && itemToAdd.canFitInsideContainerItems();
+    }
+
+    // Paper start - correct bundle inventory action
+    public int getMaxAmountToAdd(final ItemStack stack) {
+        DataResult<Fraction> weight = this.weight();
+        return weight.isError() ? 0 : Mutable.getMaxAmountToAdd(stack, weight.getOrThrow());
+    }
+    // Paper end - correct bundle inventory action
 
     public int getNumberOfItemsToShow() {
         int numberOfItemStacks = this.size();
@@ -92,8 +_,7 @@
         return Math.min(numberOfItemStacks, availableItemsToShow - emptySpaceOnNonFullRow);
     }
 
-    @Override
-    public Stream<ItemStack> itemCopies() {
+    public Stream<ItemStack> itemCopyStream() {
         return this.items.stream().map(ItemStackTemplate::create);
     }
 
@@ -101,7 +_,6 @@
         return this.items;
     }
 
-    @Override
     public int size() {
         return this.items.size();
     }
@@ -122,27 +_,6 @@
         return this.selectedItem == -1 ? null : this.items.get(this.selectedItem);
     }
 
-    public BundleContents copyWithContents(Stream<ItemStack> newContents) {
-        BundleContents.Mutable builder = new BundleContents.Mutable();
-        newContents.forEach(builder::tryInsert);
-        return builder.toImmutable();
-    }
-
-    public BundleContents.Mutable asMutable() {
-        DataResult<Fraction> currentWeight = this.weight.get();
-        if (currentWeight.isError()) {
-            return new BundleContents.Mutable();
-        }
-
-        List<ItemStack> itemsList = new ArrayList<>(this.items.size());
-
-        for (ItemStackTemplate item : this.items) {
-            itemsList.add(item.create());
-        }
-
-        return new BundleContents.Mutable(itemsList, currentWeight.getOrThrow(), this.selectedItem);
-    }
-
     @Override
     public boolean equals(Object obj) {
         if (this == obj) {
@@ -162,19 +_,27 @@
         return "BundleContents" + this.items;
     }
 
-    public static class Mutable extends SimpleMutableContainer<BundleContents> {
+    public static class Mutable {
+        private final List<ItemStack> items;
         private Fraction weight;
         private int selectedItem;
-        private boolean needsFlattening;
-
-        private Mutable(List<ItemStack> items, Fraction weight, int selectedItem) {
-            super(items);
-            this.weight = weight;
-            this.selectedItem = selectedItem;
-        }
-
-        public Mutable() {
-            this(new ArrayList<>(), Fraction.ZERO, -1);
+
+        public Mutable(BundleContents contents) {
+            DataResult<Fraction> currentWeight = contents.weight.get();
+            if (currentWeight.isError()) {
+                this.items = new ArrayList<>();
+                this.weight = Fraction.ZERO;
+                this.selectedItem = -1;
+            } else {
+                this.items = new ArrayList<>(contents.items.size());
+
+                for (ItemStackTemplate item : contents.items) {
+                    this.items.add(item.create());
+                }
+
+                this.weight = currentWeight.getOrThrow();
+                this.selectedItem = contents.selectedItem;
+            }
         }
 
         public BundleContents.Mutable clearItems() {
@@ -184,15 +_,12 @@
             return this;
         }
 
-        private int findStackIndexWithinRange(ItemStack itemsToAdd, int minInclusive, int maxExclusive) {
+        private int findStackIndex(ItemStack itemsToAdd) {
             if (!itemsToAdd.isStackable()) {
                 return -1;
             }
 
-            int startIndex = Math.max(minInclusive, 0);
-            int endIndex = Math.min(maxExclusive, this.items.size());
-
-            for (int i = startIndex; i < endIndex; i++) {
+            for (int i = 0; i < this.items.size(); i++) {
                 if (ItemStack.isSameItemSameComponents(this.items.get(i), itemsToAdd)) {
                     return i;
                 }
@@ -201,15 +_,19 @@
             return -1;
         }
 
-        private int findStackIndex(ItemStack itemsToAdd) {
-            return this.findStackIndexWithinRange(itemsToAdd, 0, this.items.size());
-        }
-
         private int getMaxAmountToAdd(Fraction itemWeight) {
             Fraction remainingWeight = Fraction.ONE.subtract(this.weight);
             return Math.max(remainingWeight.divideBy(itemWeight).intValue(), 0);
         }
 
+        // Paper start - correct bundle inventory action
+        // Static overload to easily compute this value without the need for an instance of mutable.
+        static int getMaxAmountToAdd(final ItemStack stack, final Fraction weight) {
+            Fraction remainingWeight = Fraction.ONE.subtract(weight);
+            return Math.max(remainingWeight.divideBy(BundleContents.getWeight(stack).getOrThrow()).intValue(), 0);
+        }
+        // Paper end - correct bundle inventory action
+
         public int tryInsert(ItemStack itemsToAdd) {
             if (!BundleContents.canItemBeInBundle(itemsToAdd)) {
                 return 0;
@@ -226,15 +_,15 @@
                 return 0;
             }
 
-            this.weight = this.weight.add(getStackedWeight(itemWeight, amountToAdd));
+            this.weight = this.weight.add(itemWeight.multiplyBy(Fraction.getFraction(amountToAdd, 1)));
             int stackIndex = this.findStackIndex(itemsToAdd);
             if (stackIndex != -1) {
                 ItemStack removedStack = this.items.remove(stackIndex);
                 ItemStack mergedStack = removedStack.copyWithCount(removedStack.getCount() + amountToAdd);
                 itemsToAdd.shrink(amountToAdd);
-                this.items.addFirst(mergedStack);
+                this.items.add(0, mergedStack);
             } else {
-                this.items.addFirst(itemsToAdd.split(amountToAdd));
+                this.items.add(0, itemsToAdd.split(amountToAdd));
             }
 
             return amountToAdd;
@@ -266,124 +_,16 @@
 
             int removeIndex = this.indexIsOutsideAllowedBounds(this.selectedItem) ? 0 : this.selectedItem;
             ItemStack stack = this.items.remove(removeIndex).copy();
-            this.weight = this.weight.subtract(getStackedWeight(stack));
+            this.weight = this.weight.subtract(BundleContents.getWeight(stack).getOrThrow().multiplyBy(Fraction.getFraction(stack.getCount(), 1)));
             this.toggleSelectedItem(-1);
             return stack;
         }
 
-        private static Fraction getStackedWeight(Fraction weight, int count) {
-            return weight.multiplyBy(Fraction.getFraction(count, 1));
-        }
-
-        private static Fraction getStackedWeight(ItemStack stack) {
-            return getStackedWeight(BundleContents.getWeight(stack).getOrThrow(), stack.getCount());
-        }
-
         public Fraction weight() {
             return this.weight;
         }
 
-        @Override
-        public int replaceSlotItems(ItemProvider newItems, SlotSelector slotSelector) {
-            this.mergeIdenticalStacks();
-            return super.replaceSlotItems(newItems, slotSelector);
-        }
-
-        @Override
-        public void modifySlots(Consumer<? super SlotAccess> consumer, SlotSelector slotSelector) {
-            this.mergeIdenticalStacks();
-            super.modifySlots(consumer, slotSelector);
-        }
-
-        @Override
-        protected int insertNewSlots(ItemProvider newItems, SlotSelector slotSelector) {
-            int successCount;
-            for (successCount = 0; newItems.hasNext() && this.canInsertNewSlots() && slotSelector.trySelectSlot(ItemStack.EMPTY); successCount++) {
-                boolean success = newItems.findNextNonEmpty() && this.addSlotWithItem(newItems.peek());
-                if (!success) {
-                    break;
-                }
-
-                newItems.next();
-            }
-
-            return successCount;
-        }
-
-        @Override
-        protected boolean setItem(int slot, ItemStack itemStack) {
-            ItemStack currentItem = this.items.get(slot);
-            Fraction adjustedWeight = currentItem.isEmpty() ? this.weight : this.weight.subtract(getStackedWeight(currentItem));
-            Fraction newWeight = itemStack.isEmpty() ? adjustedWeight : getWeightWithAddedItems(adjustedWeight, itemStack);
-            if (newWeight != null && super.setItem(slot, itemStack)) {
-                this.weight = newWeight;
-                this.needsFlattening = true;
-                return true;
-            } else {
-                return false;
-            }
-        }
-
-        @Override
-        protected boolean addSlotWithItem(ItemStack itemStack) {
-            if (itemStack.isEmpty()) {
-                return false;
-            } else {
-                Fraction newWeight = getWeightWithAddedItems(this.weight, itemStack);
-                if (newWeight != null && super.addSlotWithItem(itemStack)) {
-                    this.weight = newWeight;
-                    this.needsFlattening = true;
-                    return true;
-                } else {
-                    return false;
-                }
-            }
-        }
-
-        private static @Nullable Fraction getWeightWithAddedItems(Fraction weight, ItemStack itemsToAdd) {
-            if (BundleContents.canItemBeInBundle(itemsToAdd)) {
-                DataResult<Fraction> itemWeight = BundleContents.getWeight(itemsToAdd);
-                if (itemWeight.isError()) {
-                    return null;
-                }
-
-                Fraction newWeight = weight.add(getStackedWeight(itemWeight.getOrThrow(), itemsToAdd.getCount()));
-                if (newWeight.compareTo(Fraction.ONE) <= 0) {
-                    return newWeight;
-                }
-            }
-
-            return null;
-        }
-
-        @Override
-        public boolean canInsertNewSlots() {
-            return this.weight.compareTo(Fraction.ONE) < 0;
-        }
-
-        private void mergeIdenticalStacks() {
-            if (this.needsFlattening) {
-                for (int index = 0; index < this.items.size(); index++) {
-                    ItemStack itemStack = this.items.get(index);
-                    if (itemStack.isEmpty()) {
-                        this.items.remove(index--);
-                    } else {
-                        int stackIndex = this.findStackIndexWithinRange(itemStack, index + 1, this.items.size());
-                        if (stackIndex != -1) {
-                            ItemStack targetStack = this.items.get(stackIndex);
-                            int mergedCount = itemStack.getCount() + targetStack.getCount();
-                            this.items.set(stackIndex, itemStack.copyWithCount(mergedCount));
-                            this.items.remove(index--);
-                        }
-                    }
-                }
-
-                this.needsFlattening = false;
-            }
-        }
-
         public BundleContents toImmutable() {
-            this.mergeIdenticalStacks();
             Builder<ItemStackTemplate> builder = ImmutableList.builder();
 
             for (ItemStack item : this.items) {
