--- a/net/minecraft/world/item/component/Compostable.java
+++ b/net/minecraft/world/item/component/Compostable.java
@@ -3,20 +_,16 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import io.netty.buffer.ByteBuf;
+import net.minecraft.core.registries.Registries;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.resources.ResourceKey;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ResolvableInt;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
 
-public record Compostable(ResolvableInt layers) {
+public record Compostable(ResourceKey<NumberProvider> layers) {
     public static final Codec<Compostable> CODEC = RecordCodecBuilder.create(
-        i -> i.group(ResolvableInt.CODEC.fieldOf("layers").forGetter(Compostable::layers)).apply(i, Compostable::new)
+        i -> i.group(ResourceKey.codec(Registries.NUMBER_PROVIDER).fieldOf("layers").forGetter(Compostable::layers)).apply(i, Compostable::new)
     );
     public static final StreamCodec<ByteBuf, Compostable> STREAM_CODEC = StreamCodec.composite(
-        ResolvableInt.STREAM_CODEC, Compostable::layers, Compostable::new
+        ResourceKey.streamCodec(Registries.NUMBER_PROVIDER), Compostable::layers, Compostable::new
     );
-
-    public Compostable(ResourceKey<ContextIntProvider> layers) {
-        this(ResolvableInt.fromKey(layers));
-    }
 }
