--- a/net/minecraft/world/item/component/ConsumableListener.java
+++ b/net/minecraft/world/item/component/ConsumableListener.java
@@ -6,4 +_,6 @@
 
 public interface ConsumableListener {
     void onConsume(final Level level, final LivingEntity user, final ItemStack stack, final Consumable consumable);
+
+    default void cancelUsingItem(net.minecraft.server.level.ServerPlayer serverplayer, ItemStack itemstack) {} // CraftBukkit
 }
