--- a/net/minecraft/world/item/component/DeathProtection.java
+++ b/net/minecraft/world/item/component/DeathProtection.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.item.consume_effects.ApplyStatusEffectsConsumeEffect;
 import net.minecraft.world.item.consume_effects.ClearAllStatusEffectsConsumeEffect;
 import net.minecraft.world.item.consume_effects.ConsumeEffect;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
 
 public record DeathProtection(List<ConsumeEffect> deathEffects) {
     public static final Codec<DeathProtection> CODEC = RecordCodecBuilder.create(
@@ -37,7 +_,7 @@
 
     public void applyEffects(ItemStack itemStack, LivingEntity entity) {
         for (ConsumeEffect effect : this.deathEffects) {
-            effect.apply(entity.level(), itemStack, entity);
+            effect.apply(entity.level(), itemStack, entity, EntityPotionEffectEvent.Cause.TOTEM); // CraftBukkit
         }
     }
 }
