--- a/net/minecraft/world/item/component/Fireworks.java
+++ b/net/minecraft/world/item/component/Fireworks.java
@@ -36,7 +_,7 @@
     @Override
     public void addToTooltip(Item.TooltipContext context, Consumer<Component> consumer, TooltipFlag flag, DataComponentGetter components) {
         if (this.flightDuration > 0) {
-            consumer.accept(Component.translatable("item.minecraft.firework_rocket.flight_duration", this.flightDuration).withStyle(ChatFormatting.GRAY));
+            consumer.accept(Component.translatable("item.minecraft.firework_rocket.flight", this.flightDuration).withStyle(ChatFormatting.GRAY));
         }
 
         FireworkExplosion current = null;
