--- a/net/minecraft/world/item/component/PiercingWeapon.java
+++ b/net/minecraft/world/item/component/PiercingWeapon.java
@@ -75,7 +_,6 @@
     public void attack(LivingEntity attacker, EquipmentSlot hand) {
         float damage = (float)attacker.getAttributeValue(Attributes.ATTACK_DAMAGE);
         ItemStack weaponItem = attacker.getItemBySlot(hand);
-        SwingAnimation swingAnimation = weaponItem.getAttackAnimation();
         AttackRange attackRange = attacker.getAttackRangeWith(weaponItem);
         boolean hitSomething = false;
 
@@ -91,6 +_,6 @@
         }
 
         this.makeSound(attacker);
-        attacker.swingAndResetAttackStrength(InteractionHand.MAIN_HAND, swingAnimation, false);
+        attacker.swing(InteractionHand.MAIN_HAND, false);
     }
 }
