--- a/net/minecraft/world/item/component/ResolvableProfile.java
+++ b/net/minecraft/world/item/component/ResolvableProfile.java
@@ -121,7 +_,7 @@
         }
 
         @Override
-        protected Either<GameProfile, ResolvableProfile.Partial> unpack() {
+        public Either<GameProfile, ResolvableProfile.Partial> unpack() {
             return Either.right(new ResolvableProfile.Partial(this.nameOrId.left(), this.nameOrId.right(), PropertyMap.EMPTY));
         }
 
@@ -176,7 +_,7 @@
         }
 
         @Override
-        protected Either<GameProfile, ResolvableProfile.Partial> unpack() {
+        public Either<GameProfile, ResolvableProfile.Partial> unpack() {
             return this.contents;
         }
 
