--- a/net/minecraft/world/item/component/SwingAnimation.java
+++ b/net/minecraft/world/item/component/SwingAnimation.java
@@ -13,7 +_,7 @@
     public static final Codec<SwingAnimation> CODEC = RecordCodecBuilder.create(
         i -> i.group(
                 SwingAnimationType.CODEC.optionalFieldOf("type", DEFAULT.type).forGetter(SwingAnimation::type),
-                ExtraCodecs.NON_NEGATIVE_INT.optionalFieldOf("duration", DEFAULT.duration).forGetter(SwingAnimation::duration)
+                ExtraCodecs.POSITIVE_INT.optionalFieldOf("duration", DEFAULT.duration).forGetter(SwingAnimation::duration)
             )
             .apply(i, SwingAnimation::new)
     );
