--- a/net/minecraft/world/item/component/TypedEntityData.java
+++ b/net/minecraft/world/item/component/TypedEntityData.java
@@ -11,7 +_,6 @@
 import java.util.function.Consumer;
 import java.util.function.Function;
 import net.minecraft.ChatFormatting;
-import net.minecraft.core.DefaultedRegistry;
 import net.minecraft.core.HolderLookup;
 import net.minecraft.core.component.DataComponentGetter;
 import net.minecraft.nbt.CompoundTag;
@@ -26,7 +_,6 @@
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.TooltipFlag;
-import net.minecraft.world.level.SpawnData;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.storage.TagValueInput;
 import net.minecraft.world.level.storage.TagValueOutput;
@@ -85,6 +_,32 @@
         this.tag = stripId(data);
     }
 
+    // Paper start - utils for item meta
+    public static <IdType> TypedEntityData<IdType> decode(final Codec<IdType> idTypeCodec, final CompoundTag tag) {
+        return codec(idTypeCodec).decode(net.minecraft.nbt.NbtOps.INSTANCE, tag).result().orElseThrow().getFirst();
+    }
+
+    public static TypedEntityData<EntityType<?>> decodeEntity(final CompoundTag tag) {
+        return decode(net.minecraft.world.entity.EntityType.CODEC, tag);
+    }
+
+    public static TypedEntityData<net.minecraft.world.level.block.entity.BlockEntityType<?>> decodeBlockEntity(final CompoundTag tag) {
+        return decode(net.minecraft.core.registries.BuiltInRegistries.BLOCK_ENTITY_TYPE.byNameCodec(), tag);
+    }
+
+    public CompoundTag copyTagWithEntityId() {
+        final CompoundTag tag = this.tag.copy();
+        tag.putString("id", EntityType.getKey((EntityType<?>) this.type).toString());
+        return tag;
+    }
+
+    public CompoundTag copyTagWithBlockEntityId() {
+        final CompoundTag tag = this.tag.copy();
+        tag.putString("id", net.minecraft.core.registries.BuiltInRegistries.BLOCK_ENTITY_TYPE.getKey((net.minecraft.world.level.block.entity.BlockEntityType<?>) this.type).toString());
+        return tag;
+    }
+    // Paper end - utils for item meta
+
     public static <T> TypedEntityData<T> of(T type, CompoundTag data) {
         return new TypedEntityData<>(type, data);
     }
@@ -136,14 +_,6 @@
             entity.load(TagValueInput.create(reporter, entity.registryAccess(), entityData));
             entity.setUUID(uuid);
         }
-    }
-
-    public boolean loadInto(SpawnData spawnData, DefaultedRegistry<IdType> registry) {
-        CompoundTag entityTag = spawnData.getEntityToSpawn();
-        CompoundTag oldTag = entityTag.copy();
-        spawnData.getEntityToSpawn().putString("id", registry.getKey(this.type()).toString());
-        entityTag.merge(this.getUnsafe());
-        return !entityTag.equals(oldTag);
     }
 
     public boolean loadInto(BlockEntity blockEntity, HolderLookup.Provider registries) {
