--- a/net/minecraft/world/entity/monster/warden/WardenSpawnTracker.java
+++ b/net/minecraft/world/entity/monster/warden/WardenSpawnTracker.java
@@ -32,9 +_,9 @@
     private static final int WARNING_CHECK_DIAMETER = 48;
     private static final int DECREASE_WARNING_LEVEL_EVERY_INTERVAL = 12000;
     private static final int WARNING_LEVEL_INCREASE_COOLDOWN = 200;
-    private int ticksSinceLastWarning;
+    public int ticksSinceLastWarning;
     private int warningLevel;
-    private int cooldownTicks;
+    public int cooldownTicks;
 
     public WardenSpawnTracker(int p_219568_, int p_219569_, int p_219570_) {
         this.ticksSinceLastWarning = p_219568_;
@@ -103,7 +_,7 @@
         return p_219595_.getPlayers(predicate.and(LivingEntity::isAlive).and(EntitySelector.NO_SPECTATORS));
     }
 
-    private void increaseWarningLevel() {
+    public void increaseWarningLevel() {
         if (!this.onCooldown()) {
             this.ticksSinceLastWarning = 0;
             this.cooldownTicks = 200;
