--- a/net/minecraft/client/renderer/feature/MovingBlockFeatureRenderer.java
+++ b/net/minecraft/client/renderer/feature/MovingBlockFeatureRenderer.java
@@ -36,24 +_,10 @@
             this.poseStack.setIdentity();
             this.poseStack.mulPose(submit.pose());
             BlockQuadOutput quadOutput = (x, y, z, quad, instance) -> this.putBakedQuad(
-                this.poseStack,
-                x,
-                y,
-                z,
-                quad,
-                instance,
-                submit.forceTranslucent() ? ChunkSectionLayer.TRANSLUCENT : quad.materialInfo().layer(),
-                submit.outlineColor()
+                this.poseStack, x, y, z, quad, instance, quad.materialInfo().layer(), submit.outlineColor()
             );
             BlockQuadOutput solidQuadOutput = (x, y, z, quad, instance) -> this.putBakedQuad(
-                this.poseStack,
-                x,
-                y,
-                z,
-                quad,
-                instance,
-                submit.forceTranslucent() ? ChunkSectionLayer.TRANSLUCENT : ChunkSectionLayer.SOLID,
-                submit.outlineColor()
+                this.poseStack, x, y, z, quad, instance, ChunkSectionLayer.SOLID, submit.outlineColor()
             );
             BlockQuadOutput blockOutput = ModelBlockRenderer.forceOpaque(cutoutLeaves, blockState) ? solidQuadOutput : quadOutput;
             long blockSeed = blockState.getSeed(movingBlockRenderState.randomSeedPos);
@@ -82,8 +_,7 @@
         poseStack.popPose();
     }
 
-    public record Submit(Matrix4fc pose, MovingBlockRenderState movingBlockRenderState, int outlineColor, boolean forceTranslucent)
-        implements TranslucentSubmit {
+    public record Submit(Matrix4fc pose, MovingBlockRenderState movingBlockRenderState, int outlineColor) implements TranslucentSubmit {
         @Override
         public float distanceToCameraSq() {
             return TranslucentSubmit.computeDistanceToCameraSq(this.pose, 0.5F, 0.5F, 0.5F);
