--- a/net/minecraft/client/renderer/feature/QuadParticleFeatureRenderer.java
+++ b/net/minecraft/client/renderer/feature/QuadParticleFeatureRenderer.java
@@ -75,7 +_,7 @@
         renderPass.pushDebugGroup(() -> "Particles - " + (group.translucent ? "Translucent" : "Solid"));
         RenderSystem.bindDefaultUniforms(renderPass);
         renderPass.setUniform("DynamicTransforms", Objects.requireNonNull(this.dynamicTransforms));
-        renderPass.setUniform("Sampler2", context.lightmap(), RenderSystem.getSamplerCache().getClampToEdge(FilterMode.LINEAR));
+        renderPass.bindTexture("Sampler2", context.lightmap(), RenderSystem.getSamplerCache().getClampToEdge(FilterMode.LINEAR));
         drawLayers(context.stagedVertexBuffer(), group, renderPass, stage);
         renderPass.popDebugGroup();
     }
@@ -90,7 +_,7 @@
                 renderPass.setVertexBuffer(0, executeInfo.vertexBuffer().slice());
                 renderPass.setIndexBuffer(executeInfo.indexBuffer(), executeInfo.indexType());
                 AbstractTexture texture = group.textures.get(entry.getKey());
-                renderPass.setUniform("Sampler0", texture.getTextureView(), texture.getSampler());
+                renderPass.bindTexture("Sampler0", texture.getTextureView(), texture.getSampler());
                 renderPass.drawIndexed(executeInfo.indexCount(), 1, executeInfo.firstIndex(), executeInfo.baseVertex(), 0);
             }
         }
