--- a/net/minecraft/world/level/levelgen/structure/templatesystem/BlockRotProcessor.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/BlockRotProcessor.java
@@ -6,8 +_,8 @@
 import java.util.Optional;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.Block;
@@ -16,7 +_,7 @@
 public class BlockRotProcessor implements StructureProcessor {
     public static final MapCodec<BlockRotProcessor> MAP_CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(
-                RegistryCodecs.holderSet(Registries.BLOCK).optionalFieldOf("rottable_blocks").forGetter(t -> t.rottableBlocks),
+                RegistryCodecs.homogeneousList(Registries.BLOCK).optionalFieldOf("rottable_blocks").forGetter(t -> t.rottableBlocks),
                 Codec.floatRange(0.0F, 1.0F).fieldOf("integrity").forGetter(t -> t.integrity)
             )
             .apply(i, BlockRotProcessor::new)
