--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplateManager.java
@@ -96,7 +_,7 @@
         return this.sources.stream().flatMap(TemplateSource::list).distinct();
     }
 
-    public Optional<StructureTemplate> tryLoad(Identifier id) {
+    private Optional<StructureTemplate> tryLoad(Identifier id) {
         for (TemplateSource source : this.sources) {
             try {
                 Optional<StructureTemplate> loaded = source.load(id);
