--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
@@ -21,7 +_,7 @@
     private LiquidSettings liquidSettings = LiquidSettings.APPLY_WATERLOGGING;
     @Nullable
     private RandomSource random;
-    public int palette;
+    private int palette;
     private final List<StructureProcessor> processors = Lists.newArrayList();
     private boolean knownShape;
     private boolean finalizeEntities;
