--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -27,6 +_,7 @@
 import net.minecraft.nbt.IntTag;
 import net.minecraft.nbt.ListTag;
 import net.minecraft.nbt.NbtUtils;
+import net.minecraft.nbt.Tag;
 import net.minecraft.resources.Identifier;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.util.ProblemReporter;
@@ -61,6 +_,8 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.BitSetDiscreteVoxelShape;
 import net.minecraft.world.phys.shapes.DiscreteVoxelShape;
+import org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer;
+import org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -82,6 +_,11 @@
     private Vec3i size = Vec3i.ZERO;
     private String author = "?";
 
+    // CraftBukkit start - data containers
+    private static final CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new CraftPersistentDataTypeRegistry();
+    public CraftPersistentDataContainer persistentDataContainer = new CraftPersistentDataContainer(DATA_TYPE_REGISTRY);
+    // CraftBukkit end
+
     public Vec3i getSize() {
         return this.size;
     }
@@ -109,20 +_,24 @@
             this.size = size;
 
             try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(LOGGER)) {
-                level.findBlocksIn(minCorner, maxCorner).filterState(state -> !ignoreBlocks.contains(state.getBlock())).forEach((pos, state) -> {
+                for (BlockPos pos : BlockPos.betweenClosed(minCorner, maxCorner)) {
                     BlockPos relativePos = pos.subtract(minCorner);
-                    BlockEntity blockEntity = level.getBlockEntity(pos);
-                    StructureTemplate.StructureBlockInfo info;
-                    if (blockEntity != null) {
-                        TagValueOutput output = TagValueOutput.createWithContext(reporter, level.registryAccess());
-                        blockEntity.saveWithId(output);
-                        info = new StructureTemplate.StructureBlockInfo(relativePos, state, output.buildResult());
-                    } else {
-                        info = new StructureTemplate.StructureBlockInfo(relativePos, state, null);
+                    BlockState blockState = level.getBlockState(pos);
+                    if (!ignoreBlocks.stream().anyMatch(blockState::is)) {
+                        BlockEntity blockEntity = level.getBlockEntity(pos);
+                        StructureTemplate.StructureBlockInfo info;
+                        if (blockEntity != null) {
+                            TagValueOutput output = TagValueOutput.createWithContext(reporter, level.registryAccess());
+                            blockEntity.saveWithId(output);
+                            info = new StructureTemplate.StructureBlockInfo(relativePos, blockState, output.buildResult());
+                        } else {
+                            info = new StructureTemplate.StructureBlockInfo(relativePos, blockState, null);
+                        }
+
+                        addToLists(info, fullBlockList, blockEntitiesList, otherBlocksList);
                     }
+                }
 
-                    addToLists(info, fullBlockList, blockEntitiesList, otherBlocksList);
-                });
                 List<StructureTemplate.StructureBlockInfo> blockInfoList = buildInfoList(fullBlockList, blockEntitiesList, otherBlocksList);
                 this.palettes.clear();
                 this.palettes.add(new StructureTemplate.Palette(blockInfoList));
@@ -205,7 +_,7 @@
             result.add(
                 jigsaw.withInfo(
                     new StructureTemplate.StructureBlockInfo(
-                        calculateRelativePosition(settings, blockInfo.pos()).offset(position), blockInfo.state.rotate(settings.getRotation()), blockInfo.nbt
+                        calculateRelativePosition(settings, blockInfo.pos()).offset(position), blockInfo.state.rotate(settings.getRotation()), net.neoforged.neoforge.common.CommonHooks.captureStructureRotation(blockInfo.nbt, rotation)
                     )
                 )
             );
@@ -224,7 +_,7 @@
         for (StructureTemplate.StructureBlockInfo blockInfo : settings.getRandomPalette(this.palettes, position).blocks(block)) {
             BlockPos blockPos = absolute ? calculateRelativePosition(settings, blockInfo.pos).offset(position) : blockInfo.pos;
             if (boundingBox == null || boundingBox.isInside(blockPos)) {
-                result.add(new StructureTemplate.StructureBlockInfo(blockPos, blockInfo.state.rotate(settings.getRotation()), blockInfo.nbt));
+                result.add(new StructureTemplate.StructureBlockInfo(blockPos, blockInfo.state.rotate(settings.getRotation()), net.neoforged.neoforge.common.CommonHooks.captureStructureRotation(blockInfo.nbt, settings.getRotation())));
             }
         }
 
@@ -241,6 +_,10 @@
         return transform(pos, settings.getMirror(), settings.getRotation(), settings.getRotationPivot());
     }
 
+    public static Vec3 transformedVec3d(StructurePlaceSettings placementIn, Vec3 pos) {
+        return transform(pos, placementIn.getMirror(), placementIn.getRotation(), placementIn.getRotationPivot());
+    }
+
     public boolean placeInWorld(
         ServerLevelAccessor level,
         BlockPos position,
@@ -252,7 +_,19 @@
         if (this.palettes.isEmpty()) {
             return false;
         }
-
+        // CraftBukkit start
+        // We only want the TransformerLevelAccessor at certain locations because in here are many "block update" calls that shouldn't be transformed
+        ServerLevelAccessor wrappedAccessor = level;
+        org.bukkit.craftbukkit.util.CraftStructureTransformer structureTransformer = null;
+        if (wrappedAccessor instanceof org.bukkit.craftbukkit.util.TransformerLevelAccessor transformerAccessor) {
+            level = transformerAccessor.getDelegate();
+            structureTransformer = transformerAccessor.getStructureTransformer();
+            // The structureTransformer is not needed if we can not transform blocks therefore we can save a little bit of performance doing this
+            if (structureTransformer != null && !structureTransformer.canTransformBlocks()) {
+                structureTransformer = null;
+            }
+        }
+        // CraftBukkit end
         List<StructureTemplate.StructureBlockInfo> blockInfoList = settings.getRandomPalette(this.palettes, position).blocks();
         if ((!blockInfoList.isEmpty() || !settings.isIgnoreEntities() && !this.entityInfoList.isEmpty())
             && this.size.getX() >= 1
@@ -268,7 +_,7 @@
             int maxX = Integer.MIN_VALUE;
             int maxY = Integer.MIN_VALUE;
             int maxZ = Integer.MIN_VALUE;
-            List<StructureTemplate.StructureBlockInfo> processedBlockInfoList = processBlockInfos(level, position, referencePos, settings, blockInfoList);
+            List<StructureTemplate.StructureBlockInfo> processedBlockInfoList = processBlockInfos(level, position, referencePos, settings, blockInfoList, this);
 
             try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(LOGGER)) {
                 for (StructureTemplate.StructureBlockInfo blockInfo : processedBlockInfoList) {
@@ -280,6 +_,20 @@
                             level.setBlock(blockPos, Blocks.BARRIER.defaultBlockState(), 820);
                         }
 
+                        // CraftBukkit start
+                        if (structureTransformer != null) {
+                            org.bukkit.craftbukkit.block.CraftBlockState craftBlockState = (org.bukkit.craftbukkit.block.CraftBlockState) org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(level, blockPos, state, null);
+                            if (blockInfo.nbt != null && craftBlockState instanceof org.bukkit.craftbukkit.block.CraftBlockEntityState<?> entityState) {
+                                entityState.loadData(blockInfo.nbt);
+                                if (craftBlockState instanceof org.bukkit.craftbukkit.block.CraftLootable<?> craftLootable) {
+                                    craftLootable.setSeed(random.nextLong());
+                                }
+                            }
+                            craftBlockState = structureTransformer.transformCraftState(craftBlockState);
+                            state = craftBlockState.getHandle();
+                            blockInfo = new StructureTemplate.StructureBlockInfo(blockPos, state, (craftBlockState instanceof org.bukkit.craftbukkit.block.CraftBlockEntityState<?> craftBlockEntityState ? craftBlockEntityState.getSnapshotNBT() : null));
+                        }
+                        // CraftBukkit end
                         if (level.setBlock(blockPos, state, updateMode)) {
                             minX = Math.min(minX, blockPos.getX());
                             minY = Math.min(minY, blockPos.getY());
@@ -291,13 +_,14 @@
                             if (blockInfo.nbt != null) {
                                 BlockEntity blockEntity = level.getBlockEntity(blockPos);
                                 if (blockEntity != null) {
-                                    if (!SharedConstants.DEBUG_STRUCTURE_EDIT_MODE && blockEntity instanceof RandomizableContainer) {
+                                    if (structureTransformer == null && !SharedConstants.DEBUG_STRUCTURE_EDIT_MODE && blockEntity instanceof RandomizableContainer) {
                                         blockInfo.nbt.putLong("LootTableSeed", random.nextLong());
                                     }
 
                                     blockEntity.loadWithComponents(
                                         TagValueInput.create(reporter.forChild(blockEntity.problemPath()), level.registryAccess(), blockInfo.nbt)
                                     );
+                                    net.neoforged.neoforge.common.CommonHooks.applyStructureRotation(blockEntity, blockInfo.nbt, settings);
                                 }
                             }
 
@@ -375,23 +_,16 @@
                         if (blockInfo.getSecond() != null) {
                             BlockEntity blockEntity = level.getBlockEntity(blockPos);
                             if (blockEntity != null) {
+                                if (!(level instanceof net.minecraft.world.level.WorldGenLevel)) { // Paper - Fix NBT pieces overriding a block entity during worldgen deadlock
                                 blockEntity.setChanged();
+                                } // Paper - Fix NBT pieces overriding a block entity during worldgen deadlock
                             }
                         }
                     }
                 }
 
                 if (!settings.isIgnoreEntities()) {
-                    this.placeEntities(
-                        level,
-                        position,
-                        settings.getMirror(),
-                        settings.getRotation(),
-                        settings.getRotationPivot(),
-                        boundingBox,
-                        settings.shouldFinalizeEntities(),
-                        reporter
-                    );
+                    this.addEntitiesToWorld(wrappedAccessor, position, settings, reporter); // CraftBukkit
                 }
             }
 
@@ -425,6 +_,10 @@
         });
     }
 
+    /**
+     * @deprecated Forge: Use {@link #processBlockInfos(ServerLevelAccessor, BlockPos, BlockPos, StructurePlaceSettings, List, StructureTemplate)} instead.
+     */
+    @Deprecated
     public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(
         ServerLevelAccessor level,
         BlockPos position,
@@ -432,6 +_,10 @@
         StructurePlaceSettings settings,
         List<StructureTemplate.StructureBlockInfo> blockInfoList
     ) {
+        return processBlockInfos(level, position, referencePos, settings, blockInfoList, null);
+    }
+
+    public static List<StructureTemplate.StructureBlockInfo> processBlockInfos(ServerLevelAccessor level, BlockPos position, BlockPos referencePos, StructurePlaceSettings settings, List<StructureTemplate.StructureBlockInfo> blockInfoList, @Nullable StructureTemplate template) {
         List<StructureTemplate.StructureBlockInfo> originalBlockInfoList = new ArrayList<>();
         List<StructureTemplate.StructureBlockInfo> processedBlockInfoList = new ArrayList<>();
         boolean processOnlyInCurrentChunk = true;
@@ -454,7 +_,7 @@
                 Iterator<StructureProcessor> iterator = settings.getProcessors().iterator();
 
                 while (processedBlockInfo != null && iterator.hasNext()) {
-                    processedBlockInfo = iterator.next().processBlock(level, position, referencePos, blockInfo.pos, processedBlockInfo, settings);
+                    processedBlockInfo = iterator.next().process(level, position, referencePos, blockInfo, processedBlockInfo, settings, template);
                 }
 
                 if (processedBlockInfo != null) {
@@ -471,39 +_,43 @@
         return processedBlockInfoList;
     }
 
-    private void placeEntities(
-        ServerLevelAccessor level,
-        BlockPos position,
-        Mirror mirror,
-        Rotation rotation,
-        BlockPos pivot,
-        @Nullable BoundingBox boundingBox,
-        boolean finalizeEntities,
-        ProblemReporter problemReporter
-    ) {
-        for (StructureTemplate.StructureEntityInfo entityInfo : this.entityInfoList) {
-            BlockPos blockPos = transform(entityInfo.blockPos, mirror, rotation, pivot).offset(position);
-            if (boundingBox == null || boundingBox.isInside(blockPos)) {
+
+    public static List<StructureTemplate.StructureEntityInfo> processEntityInfos(@Nullable StructureTemplate template, LevelAccessor p_215387_0_, BlockPos p_215387_1_, StructurePlaceSettings p_215387_2_, List<StructureTemplate.StructureEntityInfo> p_215387_3_) {
+        List<StructureTemplate.StructureEntityInfo> list = Lists.newArrayList();
+        for(StructureTemplate.StructureEntityInfo entityInfo : p_215387_3_) {
+            Vec3 pos = transformedVec3d(p_215387_2_, entityInfo.pos).add(Vec3.atLowerCornerOf(p_215387_1_));
+            BlockPos blockpos = calculateRelativePosition(p_215387_2_, entityInfo.blockPos).offset(p_215387_1_);
+            StructureTemplate.StructureEntityInfo info = new StructureTemplate.StructureEntityInfo(pos, blockpos, entityInfo.nbt);
+            for (StructureProcessor proc : p_215387_2_.getProcessors()) {
+                info = proc.processEntity(p_215387_0_, p_215387_1_, entityInfo, info, p_215387_2_, template);
+                if (info == null)
+                    break;
+            }
+            if (info != null)
+                list.add(info);
+        }
+        return list;
+    }
+
+    private void addEntitiesToWorld(ServerLevelAccessor level, BlockPos position, StructurePlaceSettings placementIn, ProblemReporter problemReporter) {
+        for(StructureTemplate.StructureEntityInfo entityInfo : processEntityInfos(this, level, position, placementIn, this.entityInfoList)) {
+            BlockPos blockpos = entityInfo.blockPos; // FORGE: Position will have already been transformed by processEntityInfos
+            if (placementIn.getBoundingBox() == null || placementIn.getBoundingBox().isInside(blockpos)) {
                 CompoundTag tag = entityInfo.nbt.copy();
-                Vec3 relativePos = transform(entityInfo.pos, mirror, rotation, pivot);
-                Vec3 pos = relativePos.add(position.getX(), position.getY(), position.getZ());
+                Vec3 pos = entityInfo.pos; // FORGE: Position will have already been transformed by processEntityInfos
                 ListTag posTag = new ListTag();
                 posTag.add(DoubleTag.valueOf(pos.x));
                 posTag.add(DoubleTag.valueOf(pos.y));
                 posTag.add(DoubleTag.valueOf(pos.z));
                 tag.put("Pos", posTag);
                 tag.remove("UUID");
-                if (tag.contains("block_pos")) {
-                    tag.store("block_pos", BlockPos.CODEC, blockPos);
-                }
-
                 createEntityIgnoreException(problemReporter, level, tag).ifPresent(entity -> {
-                    float yRot = entity.rotate(rotation);
-                    yRot += entity.mirror(mirror) - entity.getYRot();
+                    float yRot = entity.rotate(placementIn.getRotation());
+                    yRot += entity.mirror(placementIn.getMirror()) - entity.getYRot();
                     entity.snapTo(pos.x, pos.y, pos.z, yRot, entity.getXRot());
                     entity.setYBodyRot(yRot);
                     entity.setYHeadRot(yRot);
-                    if (finalizeEntities && entity instanceof Mob mob) {
+                    if (placementIn.shouldFinalizeEntities() && entity instanceof Mob mob) {
                         mob.finalizeSpawn(level, level.getCurrentDifficultyAt(BlockPos.containing(pos)), EntitySpawnReason.STRUCTURE, null);
                     }
 
@@ -514,13 +_,13 @@
     }
 
     private static Optional<Entity> createEntityIgnoreException(ProblemReporter reporter, ServerLevelAccessor level, CompoundTag tag) {
-        try {
-            return EntityType.create(
-                TagValueInput.create(reporter, level.registryAccess(), tag), level.getLevel(), new EntitySpawnRequest(EntitySpawnReason.STRUCTURE, false)
-            );
-        } catch (Exception ignored) {
-            return Optional.empty();
-        }
+        // CraftBukkit start
+        // try {
+            return EntityType.create(TagValueInput.create(reporter, level.registryAccess(), tag), level.getLevel(), new EntitySpawnRequest(EntitySpawnReason.STRUCTURE, false));
+        // } catch (Exception exception) {
+            // return Optional.empty();
+        // }
+        // CraftBukkit end
     }
 
     public Vec3i getSize(Rotation rotation) {
@@ -693,6 +_,11 @@
 
         tag.put("entities", entityList);
         tag.put("size", this.newIntegerList(this.size.getX(), this.size.getY(), this.size.getZ()));
+        // CraftBukkit start - PDC
+        if (!this.persistentDataContainer.isEmpty()) {
+            tag.put("BukkitValues", this.persistentDataContainer.toTagCompound());
+        }
+        // CraftBukkit end
         return NbtUtils.addCurrentDataVersion(tag);
     }
 
@@ -718,6 +_,11 @@
             BlockPos blockPos = new BlockPos(blockPosTag.getIntOr(0, 0), blockPosTag.getIntOr(1, 0), blockPosTag.getIntOr(2, 0));
             entityTag.getCompound("nbt").ifPresent(nbt -> this.entityInfoList.add(new StructureTemplate.StructureEntityInfo(pos, blockPos, nbt)));
         });
+        // CraftBukkit start - PDC
+        if (tag.get("BukkitValues") instanceof CompoundTag compoundTag) {
+            this.persistentDataContainer.putAll(compoundTag);
+        }
+        // CraftBukkit end
     }
 
     private void loadPalette(HolderGetter<Block> blockLookup, ListTag paletteList, ListTag blockList) {
@@ -817,7 +_,7 @@
 
     public static final class Palette {
         private final List<StructureTemplate.StructureBlockInfo> blocks;
-        private final Map<Block, List<StructureTemplate.StructureBlockInfo>> cache = Maps.newHashMap();
+        private final Map<Block, List<StructureTemplate.StructureBlockInfo>> cache = Maps.newConcurrentMap(); // Neo: Fixes MC-271899 - Make the global StructureTemplate's palette caches now thread safe for worldgen
         private @Nullable List<StructureTemplate.JigsawBlockInfo> cachedJigsaws;
 
         private Palette(List<StructureTemplate.StructureBlockInfo> blocks) {
