--- a/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
+++ b/net/minecraft/world/level/entity/PersistentEntitySectionManager.java
@@ -3,6 +_,7 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Queues;
 import com.google.common.collect.Sets;
+import com.mohistmc.youer.plugins.ban.bans.BanEntity;
 import com.mojang.logging.LogUtils;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMaps;
@@ -27,6 +_,9 @@
 import net.minecraft.util.VisibleForDebug;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.chunk.storage.EntityStorage;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRemoveEvent;
 import org.slf4j.Logger;
 
 public class PersistentEntitySectionManager<T extends EntityAccess> implements AutoCloseable {
@@ -52,6 +_,16 @@
         this.entityGetter = new LevelEntityGetterAdapter<>(this.visibleEntityStorage, this.sectionStorage);
     }
 
+    // CraftBukkit start - add method to get all entities in chunk
+    public List<Entity> getEntities(ChunkPos chunkCoordIntPair) {
+        return sectionStorage.getExistingSectionsInChunk(chunkCoordIntPair.toLong()).flatMap(EntitySection::getEntities).map(entity -> (Entity) entity).collect(Collectors.toList());
+    }
+
+    public boolean isPending(long pair) {
+        return chunkLoadStatuses.get(pair) == ChunkLoadStatus.PENDING;
+    }
+    // CraftBukkit end
+
     void removeSectionIfEmpty(long p_157510_, EntitySection<T> p_157511_) {
         if (p_157511_.isEmpty()) {
             this.sectionStorage.remove(p_157510_);
@@ -60,7 +_,7 @@
 
     private boolean addEntityUuid(T p_157558_) {
         if (!this.knownUuids.add(p_157558_.getUUID())) {
-            LOGGER.warn("UUID of added entity already exists: {}", p_157558_);
+            if (false) LOGGER.warn("UUID of added entity already exists: {}", p_157558_);
             return false;
         } else {
             return true;
@@ -71,7 +_,16 @@
         return this.addEntity(p_157534_, false);
     }
 
+    public boolean addNewEntityWithoutEvent(T entity) {
+        return this.addEntityWithoutEvent(entity, false);
+    }
+
     private boolean addEntity(T p_157539_, boolean p_157540_) {
+        if (p_157539_ instanceof Entity entity && (BanEntity.check(entity) || net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.EntityJoinLevelEvent(entity, entity.level(), p_157540_)).isCanceled())) return false;
+        return addEntityWithoutEvent(p_157539_, p_157540_);
+    }
+
+    private boolean addEntityWithoutEvent(T p_157539_, boolean p_157540_) {
         if (!this.addEntityUuid(p_157539_)) {
             return false;
         } else {
@@ -101,11 +_,17 @@
     }
 
     public void addLegacyChunkEntities(Stream<T> p_157553_) {
-        p_157553_.forEach(p_157607_ -> this.addEntity((T)p_157607_, true));
+        p_157553_.forEach(p_157607_ -> {
+            this.addEntity(p_157607_, true);
+            if (p_157607_ instanceof Entity entity) entity.onAddedToLevel();
+        });
     }
 
     public void addWorldGenChunkEntities(Stream<T> p_157560_) {
-        p_157560_.forEach(p_157605_ -> this.addEntity((T)p_157605_, false));
+        p_157560_.forEach(p_157605_ -> {
+            this.addEntity(p_157605_, false);
+            if (p_157605_ instanceof Entity entity) entity.onAddedToLevel();
+        });
     }
 
     void startTicking(T p_157565_) {
@@ -172,6 +_,12 @@
     }
 
     private boolean storeChunkSections(long p_157513_, Consumer<T> p_157514_) {
+        // CraftBukkit start - add boolean for event call
+        return storeChunkSections(p_157513_, p_157514_, false);
+    }
+
+    private boolean storeChunkSections(long p_157513_, Consumer<T> p_157514_, boolean callEvent) {
+        // CraftBukkit end
         PersistentEntitySectionManager.ChunkLoadStatus persistententitysectionmanager$chunkloadstatus = this.chunkLoadStatuses.get(p_157513_);
         if (persistententitysectionmanager$chunkloadstatus == PersistentEntitySectionManager.ChunkLoadStatus.PENDING) {
             return false;
@@ -182,6 +_,7 @@
                 .collect(Collectors.toList());
             if (list.isEmpty()) {
                 if (persistententitysectionmanager$chunkloadstatus == PersistentEntitySectionManager.ChunkLoadStatus.LOADED) {
+                    if (callEvent) CraftEventFactory.callEntitiesUnloadEvent(((EntityStorage) permanentStorage).level, new ChunkPos(p_157513_), ImmutableList.of()); // CraftBukkit
                     this.permanentStorage.storeEntities(new ChunkEntities<>(new ChunkPos(p_157513_), ImmutableList.of()));
                 }
 
@@ -190,6 +_,7 @@
                 this.requestChunkLoad(p_157513_);
                 return false;
             } else {
+                if (callEvent) CraftEventFactory.callEntitiesUnloadEvent(((EntityStorage) permanentStorage).level, new ChunkPos(p_157513_), list.stream().map(entity -> (Entity) entity).collect(Collectors.toList())); // CraftBukkit
                 this.permanentStorage.storeEntities(new ChunkEntities<>(new ChunkPos(p_157513_), list));
                 list.forEach(p_157514_);
                 return true;
@@ -207,7 +_,7 @@
     }
 
     private boolean processChunkUnload(long p_157569_) {
-        boolean flag = this.storeChunkSections(p_157569_, p_157595_ -> p_157595_.getPassengersAndSelf().forEach(this::unloadEntity));
+        boolean flag = this.storeChunkSections(p_157569_, p_157595_ -> p_157595_.getPassengersAndSelf().forEach(this::unloadEntity), true); // CraftBukkit - add boolean for event call
         if (!flag) {
             return false;
         } else {
@@ -217,6 +_,7 @@
     }
 
     private void unloadEntity(EntityAccess p_157586_) {
+        ((Entity)p_157586_).setRemoved$cause(EntityRemoveEvent.Cause.UNLOAD); // CraftBukkit - add Bukkit remove cause
         p_157586_.setRemoved(Entity.RemovalReason.UNLOADED_TO_CHUNK);
         p_157586_.setLevelCallback(EntityInLevelCallback.NULL);
     }
@@ -228,8 +_,15 @@
     private void processPendingLoads() {
         ChunkEntities<T> chunkentities;
         while ((chunkentities = this.loadingInbox.poll()) != null) {
-            chunkentities.getEntities().forEach(p_157593_ -> this.addEntity((T)p_157593_, true));
+            chunkentities.getEntities().forEach(p_157593_ -> {
+                this.addEntity(p_157593_, true);
+                if (p_157593_ instanceof Entity entity) entity.onAddedToLevel();
+            });
             this.chunkLoadStatuses.put(chunkentities.getPos().toLong(), PersistentEntitySectionManager.ChunkLoadStatus.LOADED);
+            // CraftBukkit start - call entity load event
+            List<Entity> entities = getEntities(chunkentities.getPos());
+            CraftEventFactory.callEntitiesLoadEvent(((EntityStorage) permanentStorage).level, chunkentities.getPos(), entities);
+            // CraftBukkit end
         }
     }
 
@@ -284,6 +_,14 @@
         this.permanentStorage.close();
     }
 
+    public void close(boolean save) throws IOException {
+        if (save) {
+            this.saveAll();
+        }
+        // CraftBukkit end
+        this.permanentStorage.close();
+    }
+
     public boolean isLoaded(UUID p_157551_) {
         return this.knownUuids.contains(p_157551_);
     }
@@ -367,11 +_,13 @@
 
     class Callback implements EntityInLevelCallback {
         private final T entity;
+        private final Entity realEntity;
         private long currentSectionKey;
         private EntitySection<T> currentSection;
 
         Callback(T p_157614_, long p_157615_, EntitySection<T> p_157616_) {
             this.entity = p_157614_;
+            this.realEntity = p_157614_ instanceof Entity ? (Entity) p_157614_ : null;
             this.currentSectionKey = p_157615_;
             this.currentSection = p_157616_;
         }
@@ -390,9 +_,11 @@
                 PersistentEntitySectionManager.this.removeSectionIfEmpty(this.currentSectionKey, this.currentSection);
                 EntitySection<T> entitysection = PersistentEntitySectionManager.this.sectionStorage.getOrCreateSection(i);
                 entitysection.add(this.entity);
+                long oldSectionKey = currentSectionKey;
                 this.currentSection = entitysection;
                 this.currentSectionKey = i;
                 this.updateStatus(visibility, entitysection.getStatus());
+                if (this.realEntity != null) net.neoforged.neoforge.common.CommonHooks.onEntityEnterSection(this.realEntity, oldSectionKey, i);
             }
         }
 
