--- a/net/minecraft/data/loot/packs/VanillaChestLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaChestLoot.java
@@ -1067,7 +_,6 @@
                         .add(
                             LootItem.lootTableItem(Items.COMPASS)
                                 .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
-                                .apply(SetItemDamageFunction.setDamage(UniformGenerator.between(0.15F, 0.8F)))
                                 .setWeight(1)
                         )
                         .add(LootItem.lootTableItem(Items.BUCKET).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F))).setWeight(1))
@@ -1391,7 +_,6 @@
                 )
         );
         this.spawnerLootTables(output);
-        this.abandonedCampLootTables(output);
     }
 
     public void spawnerLootTables(BiConsumer<ResourceKey<LootTable>, LootTable.Builder> output) {
@@ -2383,130 +_,5 @@
                             .apply(SetItemCountFunction.setCount(ConstantValue.exactly(2.0F)))
                     )
             );
-    }
-
-    private void abandonedCampLootTables(BiConsumer<ResourceKey<LootTable>, LootTable.Builder> output) {
-        output.accept(
-            BuiltInLootTables.ABANDONED_CAMP_BARREL,
-            LootTable.lootTable()
-                .withPool(
-                    LootPool.lootPool()
-                        .setRolls(UniformGenerator.between(4.0F, 8.0F))
-                        .add(LootItem.lootTableItem(Items.ARROW).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F))))
-                        .add(LootItem.lootTableItem(Items.BONE).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 4.0F))))
-                        .add(LootItem.lootTableItem(Items.BOWL).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F))))
-                        .add(LootItem.lootTableItem(Items.BREAD).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F))))
-                        .add(LootItem.lootTableItem(Items.COAL).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 4.0F))))
-                        .add(LootItem.lootTableItem(Items.COBWEB).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.FISHING_ROD).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.GLASS_BOTTLE).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F))))
-                        .add(LootItem.lootTableItem(Items.LEATHER).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F))))
-                        .add(LootItem.lootTableItem(Items.BUNDLE).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.RABBIT_HIDE).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 4.0F))))
-                        .add(LootItem.lootTableItem(Items.STRING).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F))))
-                        .add(LootItem.lootTableItem(Items.WHEAT).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 4.0F))))
-                        .add(
-                            LootItem.lootTableItem(Items.DYED_CANDLE.white())
-                                .setWeight(1)
-                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
-                        )
-                        .add(LootItem.lootTableItem(Items.WOODEN_AXE).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                )
-        );
-        output.accept(
-            BuiltInLootTables.ABANDONED_CAMP_COMMON_CHEST,
-            LootTable.lootTable()
-                .withPool(
-                    LootPool.lootPool()
-                        .setRolls(UniformGenerator.between(4.0F, 6.0F))
-                        .add(LootItem.lootTableItem(Items.ARROW).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(4.0F))))
-                        .add(LootItem.lootTableItem(Items.MAP).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.BONE).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 4.0F))))
-                        .add(LootItem.lootTableItem(Items.COBWEB).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.COMPASS).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.MAP).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F))))
-                        .add(
-                            LootItem.lootTableItem(Items.FIREWORK_ROCKET)
-                                .setWeight(1)
-                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 4.0F)))
-                        )
-                        .add(LootItem.lootTableItem(Items.FISHING_ROD).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.FLINT_AND_STEEL).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.GLASS_BOTTLE).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 4.0F))))
-                        .add(LootItem.lootTableItem(Items.LEAD).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F))))
-                        .add(LootItem.lootTableItem(Items.LEATHER).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 4.0F))))
-                        .add(LootItem.lootTableItem(Items.BUNDLE).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.RABBIT_HIDE).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 4.0F))))
-                        .add(LootItem.lootTableItem(Items.SADDLE).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(
-                            LootItem.lootTableItem(Items.DYED_CANDLE.white())
-                                .setWeight(1)
-                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
-                        )
-                )
-                .withPool(
-                    LootPool.lootPool()
-                        .setRolls(ConstantValue.exactly(2.0F))
-                        .add(LootItem.lootTableItem(Items.BOW).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.BUCKET).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.COPPER_AXE).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.COPPER_BOOTS).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.COPPER_CHESTPLATE).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.COPPER_LEGGINGS).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.COPPER_SPEAR).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.COPPER_SWORD).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.SPYGLASS).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.SHEARS).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                )
-        );
-        output.accept(
-            BuiltInLootTables.ABANDONED_CAMP_SECRET_CHEST,
-            LootTable.lootTable()
-                .withPool(
-                    LootPool.lootPool()
-                        .setRolls(ConstantValue.exactly(2.0F))
-                        .add(LootItem.lootTableItem(Items.DIAMOND).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(
-                            LootItem.lootTableItem(Items.POTION)
-                                .setWeight(1)
-                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
-                                .apply(SetPotionFunction.setPotion(Potions.HEALING))
-                        )
-                        .add(
-                            LootItem.lootTableItem(Items.POTION)
-                                .setWeight(1)
-                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
-                                .apply(SetPotionFunction.setPotion(Potions.LEAPING))
-                        )
-                        .add(
-                            LootItem.lootTableItem(Items.POTION)
-                                .setWeight(1)
-                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
-                                .apply(SetPotionFunction.setPotion(Potions.NIGHT_VISION))
-                        )
-                        .add(
-                            LootItem.lootTableItem(Items.POTION)
-                                .setWeight(1)
-                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
-                                .apply(SetPotionFunction.setPotion(Potions.SWIFTNESS))
-                        )
-                )
-                .withPool(
-                    LootPool.lootPool()
-                        .setRolls(UniformGenerator.between(4.0F, 6.0F))
-                        .add(LootItem.lootTableItem(Items.MAP).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.COPPER_INGOT).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F))))
-                        .add(LootItem.lootTableItem(Items.GOLD_INGOT).setWeight(1).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F))))
-                        .add(LootItem.lootTableItem(Items.IRON_INGOT).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                )
-                .withPool(
-                    LootPool.lootPool()
-                        .setRolls(UniformGenerator.between(0.0F, 1.0F))
-                        .add(LootItem.lootTableItem(Items.IRON_AXE).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.IRON_BOOTS).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.IRON_LEGGINGS).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                        .add(LootItem.lootTableItem(Items.IRON_SPEAR).setWeight(1).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
-                )
-        );
     }
 }
