--- a/net/minecraft/data/loot/packs/VanillaGiftLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaGiftLoot.java
@@ -20,7 +_,8 @@
 import net.minecraft.world.level.storage.loot.functions.SetItemCountFunction;
 import net.minecraft.world.level.storage.loot.functions.SetPotionFunction;
 import net.minecraft.world.level.storage.loot.predicates.LootItemEntityPropertyCondition;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
+import net.minecraft.world.level.storage.loot.providers.number.UniformGenerator;
 
 public class VanillaGiftLoot implements LootTableSubProvider {
     private final LootTableSubProvider.Context output;
@@ -39,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.RABBIT_HIDE).setWeight(10))
                             .add(LootItem.lootTableItem(Items.RABBIT_FOOT).setWeight(10))
                             .add(LootItem.lootTableItem(Items.CHICKEN).setWeight(10))
@@ -55,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.CHAINMAIL_HELMET))
                             .add(LootItem.lootTableItem(Items.CHAINMAIL_CHESTPLATE))
                             .add(LootItem.lootTableItem(Items.CHAINMAIL_LEGGINGS))
@@ -68,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.COOKED_RABBIT))
                             .add(LootItem.lootTableItem(Items.COOKED_CHICKEN))
                             .add(LootItem.lootTableItem(Items.COOKED_PORKCHOP))
@@ -82,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.MAP))
                             .add(LootItem.lootTableItem(Items.PAPER))
                     )
@@ -93,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.REDSTONE))
                             .add(LootItem.lootTableItem(Items.LAPIS_LAZULI))
                     )
@@ -104,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.BREAD))
                             .add(LootItem.lootTableItem(Items.PUMPKIN_PIE))
                             .add(LootItem.lootTableItem(Items.COOKIE))
@@ -116,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.COD))
                             .add(LootItem.lootTableItem(Items.SALMON))
                     )
@@ -127,71 +_,71 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.ARROW).setWeight(26))
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.SWIFTNESS))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.SLOWNESS))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.STRENGTH))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.HEALING))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.HARMING))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.LEAPING))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.REGENERATION))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.FIRE_RESISTANCE))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.WATER_BREATHING))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.INVISIBILITY))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.NIGHT_VISION))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.WEAKNESS))
                             )
                             .add(
                                 LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
+                                    .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
                                     .apply(SetPotionFunction.setPotion(Potions.POISON))
                             )
                     )
@@ -199,23 +_,23 @@
         this.output
             .accept(
                 BuiltInLootTables.LEATHERWORKER_GIFT,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.LEATHER)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.LEATHER)))
             );
         this.output
             .accept(
                 BuiltInLootTables.LIBRARIAN_GIFT,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.BOOK)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.BOOK)))
             );
         this.output
             .accept(
                 BuiltInLootTables.MASON_GIFT,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.CLAY)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.CLAY)))
             );
         this.output
             .accept(
                 BuiltInLootTables.SHEPHERD_GIFT,
                 LootTable.lootTable()
-                    .withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).addAll(Items.WOOL.map(LootItem::lootTableItem).asList()))
+                    .withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).addAll(Items.WOOL.map(LootItem::lootTableItem).asList()))
             );
         this.output
             .accept(
@@ -223,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.STONE_PICKAXE))
                             .add(LootItem.lootTableItem(Items.STONE_AXE))
                             .add(LootItem.lootTableItem(Items.STONE_HOE))
@@ -236,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.STONE_AXE))
                             .add(LootItem.lootTableItem(Items.GOLDEN_AXE))
                             .add(LootItem.lootTableItem(Items.IRON_AXE))
@@ -245,12 +_,12 @@
         this.output
             .accept(
                 BuiltInLootTables.UNEMPLOYED_GIFT,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.WHEAT_SEEDS)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.WHEAT_SEEDS)))
             );
         this.output
             .accept(
                 BuiltInLootTables.BABY_VILLAGER_GIFT,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.POPPY)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.POPPY)))
             );
         this.output
             .accept(
@@ -258,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.TORCHFLOWER_SEEDS))
                             .add(LootItem.lootTableItem(Items.PITCHER_POD))
                     )
@@ -269,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.SLIME_BALL).setWeight(1))
                             .add(EmptyLootItem.emptyItem().setWeight(699))
                     )
@@ -280,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(
                                 AlternativesEntry.alternatives(
                                     LootItem.lootTableItem(Items.EGG)
@@ -326,12 +_,12 @@
         this.output
             .accept(
                 BuiltInLootTables.ARMADILLO_SHED,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.ARMADILLO_SCUTE)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.ARMADILLO_SCUTE)))
             );
         this.output
             .accept(
                 BuiltInLootTables.TURTLE_GROW,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.TURTLE_SCUTE)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.TURTLE_SCUTE)))
             );
     }
 }
