--- a/net/minecraft/data/loot/packs/VanillaChargedCreeperExplosionLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaChargedCreeperExplosionLoot.java
@@ -23,7 +_,7 @@
 import net.minecraft.world.level.storage.loot.entries.NestedLootTable;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 import net.minecraft.world.level.storage.loot.predicates.LootItemEntityPropertyCondition;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
 
 public class VanillaChargedCreeperExplosionLoot implements LootTableSubProvider {
     private static final List<VanillaChargedCreeperExplosionLoot.Entry> ENTRIES = List.of(
@@ -51,7 +_,7 @@
             Holder.Reference<LootTable> innerLootTable = this.output
                 .accept(
                     entry.lootTable,
-                    LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(entry.item)))
+                    LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(entry.item)))
                 );
             LootItemCondition.Builder predicate = LootItemEntityPropertyCondition.hasProperties(
                 LootContext.EntityTarget.THIS, EntityPredicate.Builder.entity().entityType(EntityTypePredicate.of(this.entityTypes, entry.entityType))
@@ -65,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(AlternativesEntry.alternatives(alternatives.toArray(LootPoolEntryContainer.Builder[]::new)))
                     )
             );
