--- a/net/minecraft/data/loot/packs/VanillaShearingLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaShearingLoot.java
@@ -21,7 +_,8 @@
 import net.minecraft.world.level.storage.loot.entries.NestedLootTable;
 import net.minecraft.world.level.storage.loot.functions.SetItemCountFunction;
 import net.minecraft.world.level.storage.loot.predicates.LootItemEntityPropertyCondition;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
+import net.minecraft.world.level.storage.loot.providers.number.UniformGenerator;
 
 public class VanillaShearingLoot implements LootTableSubProvider {
     private final LootTableSubProvider.Context output;
@@ -40,9 +_,9 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(2))
-                            .add(LootItem.lootTableItem(Items.BROWN_MUSHROOM).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1))))
-                            .add(LootItem.lootTableItem(Items.RED_MUSHROOM).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1))))
+                            .setRolls(ConstantValue.exactly(2.0F))
+                            .add(LootItem.lootTableItem(Items.BROWN_MUSHROOM).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
+                            .add(LootItem.lootTableItem(Items.RED_MUSHROOM).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
                     )
             );
         ColorCollection.zipApply(
@@ -51,7 +_,7 @@
             (dyedSheep, wool) -> this.output
                 .accept(
                     dyedSheep,
-                    LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.between(1, 3)).add(LootItem.lootTableItem(wool)))
+                    LootTable.lootTable().withPool(LootPool.lootPool().setRolls(UniformGenerator.between(1.0F, 3.0F)).add(LootItem.lootTableItem(wool)))
                 )
         );
         this.output
@@ -63,12 +_,12 @@
         Holder.Reference<LootTable> shearRedMooshroom = this.output
             .accept(
                 BuiltInLootTables.SHEAR_RED_MOOSHROOM,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(5)).add(LootItem.lootTableItem(Items.RED_MUSHROOM)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(5.0F)).add(LootItem.lootTableItem(Items.RED_MUSHROOM)))
             );
         Holder.Reference<LootTable> sheadBrownMooshroom = this.output
             .accept(
                 BuiltInLootTables.SHEAR_BROWN_MOOSHROOM,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(5)).add(LootItem.lootTableItem(Items.BROWN_MUSHROOM)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(5.0F)).add(LootItem.lootTableItem(Items.BROWN_MUSHROOM)))
             );
         this.output
             .accept(
@@ -101,7 +_,7 @@
         this.output
             .accept(
                 BuiltInLootTables.SHEAR_SNOW_GOLEM,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.CARVED_PUMPKIN)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.CARVED_PUMPKIN)))
             );
     }
 }
