--- a/net/minecraft/data/loot/packs/VanillaBlockLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaBlockLoot.java
@@ -37,7 +_,7 @@
 import net.minecraft.world.level.block.TntBlock;
 import net.minecraft.world.level.block.state.properties.BedPart;
 import net.minecraft.world.level.block.state.properties.DoubleBlockHalf;
-import net.minecraft.world.level.storage.loot.IntLimit;
+import net.minecraft.world.level.storage.loot.IntRange;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.LootPool;
 import net.minecraft.world.level.storage.loot.LootTable;
@@ -57,7 +_,8 @@
 import net.minecraft.world.level.storage.loot.predicates.LootItemRandomChanceCondition;
 import net.minecraft.world.level.storage.loot.predicates.MatchBlock;
 import net.minecraft.world.level.storage.loot.predicates.MatchTool;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
+import net.minecraft.world.level.storage.loot.providers.number.UniformGenerator;
 
 public class VanillaBlockLoot extends BlockLootSubProvider {
     private static final float[] JUNGLE_LEAVES_SAPLING_CHANGES = new float[]{0.025F, 0.027777778F, 0.03125F, 0.041666668F, 0.1F};
@@ -400,11 +_,6 @@
         this.dropSelf(Blocks.TARGET);
         Blocks.GLAZED_TERRACOTTA.forEach(x$0 -> this.dropSelf(x$0));
         Blocks.CONCRETE.forEach(x$0 -> this.dropSelf(x$0));
-        Blocks.CONCRETE_STAIRS.forEach(x$0 -> this.dropSelf(x$0));
-        Blocks.CONCRETE_SLAB.forEach(block -> {
-            VanillaBlockLoot rec$ = this;
-            this.add(block, x$0 -> rec$.createSlabItemTable(x$0));
-        });
         Blocks.CONCRETE_POWDER.forEach(x$0 -> this.dropSelf(x$0));
         this.dropSelf(Blocks.KELP);
         this.dropSelf(Blocks.DRIED_KELP_BLOCK);
@@ -624,11 +_,11 @@
         this.add(Blocks.HORN_CORAL_BLOCK, block -> this.createSingleItemTableWithSilkTouch(block, Blocks.DEAD_HORN_CORAL_BLOCK));
         this.add(Blocks.CRIMSON_NYLIUM, block -> this.createSingleItemTableWithSilkTouch(block, Blocks.NETHERRACK));
         this.add(Blocks.WARPED_NYLIUM, block -> this.createSingleItemTableWithSilkTouch(block, Blocks.NETHERRACK));
-        this.add(Blocks.BOOKSHELF, block -> this.createSingleItemTableWithSilkTouch(block, Items.BOOK, ContextIntProviders.exactly(3)));
-        this.add(Blocks.CLAY, block -> this.createSingleItemTableWithSilkTouch(block, Items.CLAY_BALL, ContextIntProviders.exactly(4)));
-        this.add(Blocks.ENDER_CHEST, block -> this.createSingleItemTableWithSilkTouch(block, Blocks.OBSIDIAN, ContextIntProviders.exactly(8)));
-        this.add(Blocks.SNOW_BLOCK, block -> this.createSingleItemTableWithSilkTouch(block, Items.SNOWBALL, ContextIntProviders.exactly(4)));
-        this.add(Blocks.CHORUS_PLANT, this.createSingleItemTable(Items.CHORUS_FRUIT, ContextIntProviders.between(0, 1)));
+        this.add(Blocks.BOOKSHELF, block -> this.createSingleItemTableWithSilkTouch(block, Items.BOOK, ConstantValue.exactly(3.0F)));
+        this.add(Blocks.CLAY, block -> this.createSingleItemTableWithSilkTouch(block, Items.CLAY_BALL, ConstantValue.exactly(4.0F)));
+        this.add(Blocks.ENDER_CHEST, block -> this.createSingleItemTableWithSilkTouch(block, Blocks.OBSIDIAN, ConstantValue.exactly(8.0F)));
+        this.add(Blocks.SNOW_BLOCK, block -> this.createSingleItemTableWithSilkTouch(block, Items.SNOWBALL, ConstantValue.exactly(4.0F)));
+        this.add(Blocks.CHORUS_PLANT, this.createSingleItemTable(Items.CHORUS_FRUIT, UniformGenerator.between(0.0F, 1.0F)));
         this.dropPottedContents(Blocks.POTTED_OAK_SAPLING);
         this.dropPottedContents(Blocks.POTTED_SPRUCE_SAPLING);
         this.dropPottedContents(Blocks.POTTED_BIRCH_SAPLING);
@@ -755,7 +_,7 @@
                     this.applyExplosionCondition(
                         Blocks.TNT,
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(
                                 LootItem.lootTableItem(Blocks.TNT)
                                     .when(
@@ -772,13 +_,13 @@
             block -> LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
                                 block,
                                 LootItem.lootTableItem(Items.COCOA_BEANS)
                                     .apply(
-                                        SetItemCountFunction.setCount(ContextIntProviders.exactly(3))
+                                        SetItemCountFunction.setCount(ConstantValue.exactly(3.0F))
                                             .when(
                                                 MatchBlock.blockMatches(
                                                     this.blocks, block, StatePropertiesPredicate.Builder.properties().hasProperty(CocoaBlock.AGE, 2)
@@ -794,13 +_,13 @@
             block -> LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
                                 block,
                                 LootItem.lootTableItem(Items.SHELF_MUSHROOM)
                                     .apply(
-                                        SetItemCountFunction.setCount(ContextIntProviders.exactly(2))
+                                        SetItemCountFunction.setCount(ConstantValue.exactly(2.0F))
                                             .when(
                                                 MatchBlock.blockMatches(
                                                     this.blocks, block, StatePropertiesPredicate.Builder.properties().hasProperty(ShelfMushroomBlock.AGE, 1)
@@ -816,14 +_,14 @@
             block -> LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
                                 Blocks.SEA_PICKLE,
                                 LootItem.lootTableItem(block)
                                     .apply(
                                         List.of(2, 3, 4),
-                                        count -> SetItemCountFunction.setCount(ContextIntProviders.exactly(count))
+                                        count -> SetItemCountFunction.setCount(ConstantValue.exactly(count.intValue()))
                                             .when(
                                                 MatchBlock.blockMatches(
                                                     this.blocks,
@@ -908,7 +_,7 @@
                     this.applyExplosionCondition(
                         block,
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(
                                 LootItem.lootTableItem(block)
                                     .apply(
@@ -1046,7 +_,7 @@
                                 )
                             )
                             .add(LootItem.lootTableItem(Items.SWEET_BERRIES))
-                            .apply(SetItemCountFunction.setCount(ContextIntProviders.between(2, 3)))
+                            .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 3.0F)))
                             .apply(ApplyBonusCount.addUniformBonusCount(this.enchantments.getOrThrow(Enchantments.FORTUNE)))
                     )
                     .withPool(
@@ -1057,7 +_,7 @@
                                 )
                             )
                             .add(LootItem.lootTableItem(Items.SWEET_BERRIES))
-                            .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 2)))
+                            .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F)))
                             .apply(ApplyBonusCount.addUniformBonusCount(this.enchantments.getOrThrow(Enchantments.FORTUNE)))
                     )
             )
@@ -1084,7 +_,7 @@
                 (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
                     block,
                     LootItem.lootTableItem(Items.GOLD_NUGGET)
-                        .apply(SetItemCountFunction.setCount(ContextIntProviders.between(2, 6)))
+                        .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 6.0F)))
                         .apply(ApplyBonusCount.addOreBonusCount(this.enchantments.getOrThrow(Enchantments.FORTUNE)))
                 )
             )
@@ -1102,7 +_,7 @@
             block -> this.createShearsDispatchTable(
                 block,
                 (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
-                    block, LootItem.lootTableItem(Items.STICK).apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
+                    block, LootItem.lootTableItem(Items.STICK).apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
                 )
             )
         );
@@ -1129,7 +_,7 @@
             block -> LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             (LootPoolEntryContainer.Builder<?>)((UniformContainerBase.Builder)this.applyExplosionCondition(block, LootItem.lootTableItem(block)))
                                 .when(LootItemEntityPropertyCondition.entityPresent(LootContext.EntityTarget.THIS))
@@ -1145,9 +_,9 @@
                 (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
                     block,
                     LootItem.lootTableItem(Items.GLOWSTONE_DUST)
-                        .apply(SetItemCountFunction.setCount(ContextIntProviders.between(2, 4)))
+                        .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 4.0F)))
                         .apply(ApplyBonusCount.addUniformBonusCount(this.enchantments.getOrThrow(Enchantments.FORTUNE)))
-                        .apply(LimitCount.limitCount(IntLimit.range(1, 4)))
+                        .apply(LimitCount.limitCount(IntRange.range(1, 4)))
                 )
             )
         );
@@ -1158,9 +_,9 @@
                 (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
                     block,
                     LootItem.lootTableItem(Items.MELON_SLICE)
-                        .apply(SetItemCountFunction.setCount(ContextIntProviders.between(3, 7)))
+                        .apply(SetItemCountFunction.setCount(UniformGenerator.between(3.0F, 7.0F)))
                         .apply(ApplyBonusCount.addUniformBonusCount(this.enchantments.getOrThrow(Enchantments.FORTUNE)))
-                        .apply(LimitCount.limitCount(IntLimit.upperBound(9)))
+                        .apply(LimitCount.limitCount(IntRange.upperBound(9)))
                 )
             )
         );
@@ -1173,9 +_,9 @@
                 (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
                     block,
                     LootItem.lootTableItem(Items.PRISMARINE_CRYSTALS)
-                        .apply(SetItemCountFunction.setCount(ContextIntProviders.between(2, 3)))
+                        .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 3.0F)))
                         .apply(ApplyBonusCount.addUniformBonusCount(this.enchantments.getOrThrow(Enchantments.FORTUNE)))
-                        .apply(LimitCount.limitCount(IntLimit.range(1, 5)))
+                        .apply(LimitCount.limitCount(IntRange.range(1, 5)))
                 )
             )
         );
@@ -1186,9 +_,9 @@
                 (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
                     block,
                     LootItem.lootTableItem(Items.RESIN_CLUMP)
-                        .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 3)))
+                        .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
                         .apply(ApplyBonusCount.addUniformBonusCount(this.enchantments.getOrThrow(Enchantments.FORTUNE)))
-                        .apply(LimitCount.limitCount(IntLimit.upperBound(9)))
+                        .apply(LimitCount.limitCount(IntRange.upperBound(9)))
                 )
             )
         );
@@ -1199,11 +_,11 @@
                     this.applyExplosionDecay(
                         block,
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(
                                 LootItem.lootTableItem(Items.NETHER_WART)
                                     .apply(
-                                        SetItemCountFunction.setCount(ContextIntProviders.between(2, 4))
+                                        SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 4.0F))
                                             .when(
                                                 MatchBlock.blockMatches(
                                                     this.blocks, block, StatePropertiesPredicate.Builder.properties().hasProperty(NetherWartBlock.AGE, 3)
@@ -1240,7 +_,7 @@
                                                     StatePropertiesPredicate.Builder.properties().hasProperty(SnowLayerBlock.LAYERS, layers.intValue())
                                                 )
                                             )
-                                            .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(layers)))
+                                            .apply(SetItemCountFunction.setCount(ConstantValue.exactly(layers.intValue())))
                                     )
                                     .when(this.doesNotHaveSilkTouch()),
                                 AlternativesEntry.alternatives(
@@ -1248,7 +_,7 @@
                                     layers -> layers == 8
                                         ? LootItem.lootTableItem(Blocks.SNOW_BLOCK)
                                         : LootItem.lootTableItem(Blocks.SNOW)
-                                            .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(layers)))
+                                            .apply(SetItemCountFunction.setCount(ConstantValue.exactly(layers.intValue())))
                                             .when(
                                                 MatchBlock.blockMatches(
                                                     this.blocks,
@@ -1278,7 +_,7 @@
             block -> this.createSilkTouchDispatchTable(
                 block,
                 (LootPoolEntryContainer.Builder<?>)this.applyExplosionCondition(
-                    block, LootItem.lootTableItem(Items.CHARCOAL).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(2)))
+                    block, LootItem.lootTableItem(Items.CHARCOAL).apply(SetItemCountFunction.setCount(ConstantValue.exactly(2.0F)))
                 )
             )
         );
@@ -1289,7 +_,7 @@
                 this.applyExplosionCondition(
                     block,
                     LootItem.lootTableItem(Items.GOLD_NUGGET)
-                        .apply(SetItemCountFunction.setCount(ContextIntProviders.between(2, 5)))
+                        .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 5.0F)))
                         .when(BonusLevelTableCondition.bonusLevelFlatChance(this.enchantments.getOrThrow(Enchantments.FORTUNE), 0.1F, 0.14285715F, 0.25F, 1.0F))
                         .otherwise(LootItem.lootTableItem(block))
                 )
@@ -1300,7 +_,7 @@
             block -> this.createSilkTouchDispatchTable(
                 block,
                 (LootPoolEntryContainer.Builder<?>)this.applyExplosionCondition(
-                    block, LootItem.lootTableItem(Items.SOUL_SOIL).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
+                    block, LootItem.lootTableItem(Items.SOUL_SOIL).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
                 )
             )
         );
@@ -1309,12 +_,12 @@
             block -> this.createSilkTouchDispatchTable(
                 block,
                 LootItem.lootTableItem(Items.AMETHYST_SHARD)
-                    .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(4)))
+                    .apply(SetItemCountFunction.setCount(ConstantValue.exactly(4.0F)))
                     .apply(ApplyBonusCount.addOreBonusCount(this.enchantments.getOrThrow(Enchantments.FORTUNE)))
                     .when(MatchTool.toolMatches(ItemPredicate.Builder.item().of(this.items, ItemTags.CLUSTER_MAX_HARVESTABLES)))
                     .otherwise(
                         (LootPoolEntryContainer.Builder<?>)this.applyExplosionDecay(
-                            block, LootItem.lootTableItem(Items.AMETHYST_SHARD).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(2)))
+                            block, LootItem.lootTableItem(Items.AMETHYST_SHARD).apply(SetItemCountFunction.setCount(ConstantValue.exactly(2.0F)))
                         )
                     )
             )
@@ -1382,7 +_,7 @@
         return LootTable.lootTable()
             .withPool(
                 LootPool.lootPool()
-                    .setRolls(ContextIntProviders.exactly(1))
+                    .setRolls(ConstantValue.exactly(1.0F))
                     .add(
                         DynamicLoot.dynamicEntry(DecoratedPotBlock.SHERDS_DYNAMIC_DROP_ID)
                             .when(
@@ -1425,11 +_,11 @@
                                         ? LootItem.lootTableItem(Items.PITCHER_PLANT)
                                             .when(isAge)
                                             .when(isLowerPart)
-                                            .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
+                                            .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
                                         : LootItem.lootTableItem(Items.PITCHER_POD)
                                             .when(isAge)
                                             .when(isLowerPart)
-                                            .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)));
+                                            .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)));
                                 }
                             )
                         )
@@ -1443,7 +_,7 @@
                 this.applyExplosionCondition(
                     block,
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(block)
                                 .apply(CopyComponentsFunction.copyComponentsFromBlockEntity(LootContextParams.BLOCK_ENTITY).include(DataComponents.CUSTOM_NAME))
