--- a/net/minecraft/data/loot/packs/VanillaBlockInteractLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaBlockInteractLoot.java
@@ -14,7 +_,8 @@
 import net.minecraft.world.level.storage.loot.entries.LootItem;
 import net.minecraft.world.level.storage.loot.functions.SetItemCountFunction;
 import net.minecraft.world.level.storage.loot.predicates.MatchBlock;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
+import net.minecraft.world.level.storage.loot.providers.number.UniformGenerator;
 
 public class VanillaBlockInteractLoot implements LootTableSubProvider {
     private final LootTableSubProvider.Context output;
@@ -33,14 +_,14 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
-                            .add(LootItem.lootTableItem(Items.HONEYCOMB).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(3))))
+                            .setRolls(ConstantValue.exactly(1.0F))
+                            .add(LootItem.lootTableItem(Items.HONEYCOMB).apply(SetItemCountFunction.setCount(ConstantValue.exactly(3.0F))))
                     )
             );
         this.output
             .accept(
                 BuiltInLootTables.HARVEST_CAVE_VINE,
-                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.GLOW_BERRIES)))
+                LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.GLOW_BERRIES)))
             );
         this.output
             .accept(
@@ -50,7 +_,7 @@
                         LootPool.lootPool()
                             .add(
                                 LootItem.lootTableItem(Items.SWEET_BERRIES)
-                                    .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
+                                    .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
                                     .when(
                                         MatchBlock.blockMatches(
                                             this.blocks,
@@ -62,7 +_,7 @@
                     )
                     .withPool(
                         LootPool.lootPool()
-                            .add(LootItem.lootTableItem(Items.SWEET_BERRIES).apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 2))))
+                            .add(LootItem.lootTableItem(Items.SWEET_BERRIES).apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F))))
                     )
             );
         this.output
@@ -71,8 +_,8 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
-                            .add(LootItem.lootTableItem(Items.PUMPKIN_SEEDS).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(4))))
+                            .setRolls(ConstantValue.exactly(1.0F))
+                            .add(LootItem.lootTableItem(Items.PUMPKIN_SEEDS).apply(SetItemCountFunction.setCount(ConstantValue.exactly(4.0F))))
                     )
             );
         this.output
@@ -81,8 +_,8 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
-                            .add(LootItem.lootTableItem(Items.HANGING_ROOTS).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1))))
+                            .setRolls(ConstantValue.exactly(1.0F))
+                            .add(LootItem.lootTableItem(Items.HANGING_ROOTS).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
                     )
             );
     }
