--- a/net/minecraft/data/loot/packs/VanillaEntityLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaEntityLoot.java
@@ -38,8 +_,8 @@
 import net.minecraft.world.level.storage.loot.predicates.LootItemKilledByPlayerCondition;
 import net.minecraft.world.level.storage.loot.predicates.LootItemRandomChanceCondition;
 import net.minecraft.world.level.storage.loot.predicates.LootItemRandomChanceWithEnchantedBonusCondition;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProviders;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
+import net.minecraft.world.level.storage.loot.providers.number.UniformGenerator;
 
 public class VanillaEntityLoot extends EntityLootSubProvider {
     public VanillaEntityLoot(LootTableSubProvider.Context output) {
@@ -59,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BLAZE_ROD)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                 )
@@ -73,31 +_,29 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ARROW)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BONE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(
-                                    EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)).setLimit(1)
-                                )
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)).setLimit(1))
                                 .apply(SetPotionFunction.setPotion(Potions.POISON))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
@@ -108,8 +_,8 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
-                        .add(LootItem.lootTableItem(Items.STRING).apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2))))
+                        .setRolls(ConstantValue.exactly(1.0F))
+                        .add(LootItem.lootTableItem(Items.STRING).apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F))))
                 )
         );
         this.add(EntityTypes.CAMEL, LootTable.lootTable());
@@ -118,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ROTTEN_FLESH)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(2, 3)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 3.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -131,20 +_,20 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.STRING)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.SPIDER_EYE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(-1, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(-1.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                 )
@@ -154,20 +_,20 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.FEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.CHICKEN)
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -176,12 +_,12 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.COD).apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot())))
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.BONE_MEAL))
                         .when(LootItemRandomChanceCondition.randomChance(0.05F))
                 )
@@ -191,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.COPPER_INGOT)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 3)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -204,21 +_,21 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.LEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BEEF)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 3)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -227,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.GUNPOWDER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
@@ -249,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.COD)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
                         )
                 )
@@ -263,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.LEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -276,16 +_,16 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ROTTEN_FLESH)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.COPPER_INGOT))
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                         .when(LootItemRandomChanceWithEnchantedBonusCondition.randomChanceAndLootingBoost(this.enchantments, 0.11F, 0.02F))
@@ -298,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ENDER_PEARL)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -310,14 +_,14 @@
         this.add(
             EntityTypes.EVOKER,
             LootTable.lootTable()
-                .withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.TOTEM_OF_UNDYING)))
+                .withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.TOTEM_OF_UNDYING)))
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.EMERALD)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                 )
@@ -327,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BREEZE_ROD)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(1.0F, 2.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(1.0F, 2.0F)))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                 )
@@ -343,27 +_,27 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.GHAST_TEAR)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.GUNPOWDER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.MUSIC_DISC_TEARS))
-                        .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
+                        .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
                         .when(
                             DamageSourceCondition.hasDamageSource(
                                 this.projectileDamage().direct(EntityPredicate.Builder.entity().of(this.entityTypes, EntityTypes.FIREBALL))
@@ -379,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.GLOW_INK_SAC)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 3)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -393,32 +_,32 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.PRISMARINE_SHARD)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.COD)
                                 .setWeight(2)
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
                         )
                         .add(
                             LootItem.lootTableItem(Items.PRISMARINE_CRYSTALS)
                                 .setWeight(2)
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .add(EmptyLootItem.emptyItem())
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             NestedLootTable.lootTableReference(this.lootTables.getOrThrow(BuiltInLootTables.FISHING_FISH))
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
@@ -432,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.LEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -445,20 +_,20 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ROTTEN_FLESH)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.RABBIT_FOOT)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                                 .when(
                                     LootItemEntityPropertyCondition.hasProperties(
                                         LootContext.EntityTarget.THIS,
@@ -473,7 +_,7 @@
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.IRON_INGOT))
                         .add(LootItem.lootTableItem(Items.CARROT))
                         .add(LootItem.lootTableItem(Items.POTATO).apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot())))
@@ -486,8 +_,8 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
-                        .add(LootItem.lootTableItem(Items.SADDLE).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1))))
+                        .setRolls(ConstantValue.exactly(1.0F))
+                        .add(LootItem.lootTableItem(Items.SADDLE).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
                 )
         );
         this.add(EntityTypes.ILLUSIONER, LootTable.lootTable());
@@ -496,13 +_,13 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
-                        .add(LootItem.lootTableItem(Blocks.POPPY).apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2))))
+                        .setRolls(ConstantValue.exactly(1.0F))
+                        .add(LootItem.lootTableItem(Blocks.POPPY).apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F))))
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
-                        .add(LootItem.lootTableItem(Items.IRON_INGOT).apply(SetItemCountFunction.setCount(ContextIntProviders.between(3, 5))))
+                        .setRolls(ConstantValue.exactly(1.0F))
+                        .add(LootItem.lootTableItem(Items.IRON_INGOT).apply(SetItemCountFunction.setCount(UniformGenerator.between(3.0F, 5.0F))))
                 )
         );
         this.add(
@@ -510,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.LEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -523,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.MAGMA_CREAM)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(-2, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(-2.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                                 .when(this.killedByFrog().invert())
                                 .when(
                                     LootItemEntityPropertyCondition.hasProperties(
@@ -538,17 +_,17 @@
                         )
                         .add(
                             LootItem.lootTableItem(Items.PEARLESCENT_FROGLIGHT)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
+                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
                                 .when(this.killedByFrogVariant(FrogVariants.WARM))
                         )
                         .add(
                             LootItem.lootTableItem(Items.VERDANT_FROGLIGHT)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
+                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
                                 .when(this.killedByFrogVariant(FrogVariants.COLD))
                         )
                         .add(
                             LootItem.lootTableItem(Items.OCHRE_FROGLIGHT)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
+                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
                                 .when(this.killedByFrogVariant(FrogVariants.TEMPERATE))
                         )
                 )
@@ -560,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.LEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -573,21 +_,21 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.LEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BEEF)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 3)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -596,7 +_,7 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.NAUTILUS_SHELL))
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                         .when(LootItemRandomChanceWithEnchantedBonusCondition.randomChanceAndLootingBoost(this.enchantments, 0.05F, 0.01F))
@@ -608,8 +_,8 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
-                        .add(LootItem.lootTableItem(Blocks.BAMBOO).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1))))
+                        .setRolls(ConstantValue.exactly(1.0F))
+                        .add(LootItem.lootTableItem(Blocks.BAMBOO).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
                 )
         );
         this.add(
@@ -617,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.FEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -630,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.PHANTOM_MEMBRANE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                 )
@@ -644,12 +_,12 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.PORKCHOP)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 3)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -658,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.OMINOUS_BOTTLE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
-                                .apply(SetOminousBottleAmplifierFunction.setAmplifier(ContextIntProviders.between(0, 4)))
+                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
+                                .apply(SetOminousBottleAmplifierFunction.setAmplifier(UniformGenerator.between(0.0F, 4.0F)))
                         )
                         .when(
                             LootItemEntityPropertyCondition.hasProperties(
@@ -677,19 +_,19 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.COD)
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
                                 .setWeight(3)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .add(
                             LootItem.lootTableItem(Items.SALMON)
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -698,12 +_,12 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
-                        .add(LootItem.lootTableItem(Items.PUFFERFISH).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1))))
+                        .setRolls(ConstantValue.exactly(1.0F))
+                        .add(LootItem.lootTableItem(Items.PUFFERFISH).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.BONE_MEAL))
                         .when(LootItemRandomChanceCondition.randomChance(0.05F))
                 )
@@ -713,26 +_,26 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.RABBIT_HIDE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.RABBIT)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
+                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.RABBIT_FOOT))
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                         .when(LootItemRandomChanceWithEnchantedBonusCondition.randomChanceAndLootingBoost(this.enchantments, 0.1F, 0.03F))
@@ -743,12 +_,12 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.SALMON).apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot())))
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.BONE_MEAL))
                         .when(LootItemRandomChanceCondition.randomChance(0.05F))
                 )
@@ -758,12 +_,12 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.MUTTON)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 2)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 2.0F)))
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(createSheepDispatchPool(BuiltInLootTables.SHEEP.map(this.lootTables::getOrThrow)))
@@ -778,7 +_,7 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.SHULKER_SHELL))
                         .when(LootItemRandomChanceWithEnchantedBonusCondition.randomChanceAndLootingBoost(this.enchantments, 0.5F, 0.0625F))
                 )
@@ -789,20 +_,20 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ARROW)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BONE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -811,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BONE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -824,16 +_,16 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.SLIME_BALL)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                                 .when(this.killedByFrog().invert())
                         )
                         .add(
                             LootItem.lootTableItem(Items.SLIME_BALL)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
+                                .apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F)))
                                 .when(this.killedByFrog())
                         )
                         .when(
@@ -849,8 +_,8 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
-                        .add(LootItem.lootTableItem(Items.SNOWBALL).apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 15))))
+                        .setRolls(ConstantValue.exactly(1.0F))
+                        .add(LootItem.lootTableItem(Items.SNOWBALL).apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 15.0F))))
                 )
         );
         this.add(
@@ -858,20 +_,20 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.STRING)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.SPIDER_EYE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(-1, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(-1.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                 )
@@ -881,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.INK_SAC)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 3)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -894,31 +_,29 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ARROW)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BONE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(
-                                    EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)).setLimit(1)
-                                )
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)).setLimit(1))
                                 .apply(SetPotionFunction.setPotion(Potions.SLOWNESS))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
@@ -929,31 +_,29 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ARROW)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BONE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.TIPPED_ARROW)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(
-                                    EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)).setLimit(1)
-                                )
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)).setLimit(1))
                                 .apply(SetPotionFunction.setPotion(Potions.WEAKNESS))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
@@ -964,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.STRING)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(2, 5)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 5.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -978,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.LEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -991,12 +_,12 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
-                        .add(LootItem.lootTableItem(Items.TROPICAL_FISH).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1))))
+                        .setRolls(ConstantValue.exactly(1.0F))
+                        .add(LootItem.lootTableItem(Items.TROPICAL_FISH).apply(SetItemCountFunction.setCount(ConstantValue.exactly(1.0F))))
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.BONE_MEAL))
                         .when(LootItemRandomChanceCondition.randomChance(0.05F))
                 )
@@ -1006,17 +_,17 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Blocks.SEAGRASS)
                                 .setWeight(3)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.BOWL))
                         .when(
                             DamageSourceCondition.hasDamageSource(
@@ -1029,7 +_,7 @@
         this.add(EntityTypes.VILLAGER, LootTable.lootTable());
         this.add(
             EntityTypes.WARDEN,
-            LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(LootItem.lootTableItem(Items.SCULK_CATALYST)))
+            LootTable.lootTable().withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(LootItem.lootTableItem(Items.SCULK_CATALYST)))
         );
         this.add(EntityTypes.WANDERING_TRADER, LootTable.lootTable());
         this.add(
@@ -1037,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.EMERALD)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                 )
@@ -1051,46 +_,46 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.between(1, 3))
+                        .setRolls(UniformGenerator.between(1.0F, 3.0F))
                         .add(
                             LootItem.lootTableItem(Items.GLOWSTONE_DUST)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .add(
                             LootItem.lootTableItem(Items.SUGAR)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .add(
                             LootItem.lootTableItem(Items.SPIDER_EYE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .add(
                             LootItem.lootTableItem(Items.GLASS_BOTTLE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .add(
                             LootItem.lootTableItem(Items.GUNPOWDER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .add(
                             LootItem.lootTableItem(Items.STICK)
                                 .setWeight(2)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.REDSTONE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(4, 8)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(4.0F, 8.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -1100,25 +_,25 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.COAL)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(-1, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(-1.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.BONE)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Blocks.WITHER_SKELETON_SKULL))
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                         .when(LootItemRandomChanceWithEnchantedBonusCondition.randomChanceAndLootingBoost(this.enchantments, 0.025F, 0.01F))
@@ -1130,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ROTTEN_FLESH)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(1, 3)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(1.0F, 3.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -1144,20 +_,20 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ROTTEN_FLESH)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.RED_MUSHROOM)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                                 .when(
                                     LootItemEntityPropertyCondition.hasProperties(
                                         LootContext.EntityTarget.THIS,
@@ -1172,7 +_,7 @@
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.IRON_INGOT))
                         .add(LootItem.lootTableItem(Items.CARROT))
                         .add(LootItem.lootTableItem(Items.POTATO).apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot())))
@@ -1198,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ROTTEN_FLESH)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(2, 3)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 3.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -1211,11 +_,11 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ROTTEN_FLESH)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 3)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 3.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                 )
@@ -1225,25 +_,25 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ROTTEN_FLESH)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.GOLD_NUGGET)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.GOLD_INGOT))
                         .when(LootItemKilledByPlayerCondition.killedByPlayer())
                         .when(LootItemRandomChanceWithEnchantedBonusCondition.randomChanceAndLootingBoost(this.enchantments, 0.025F, 0.01F))
@@ -1254,21 +_,21 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.PORKCHOP)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(2, 4)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(2.0F, 4.0F)))
                                 .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.LEATHER)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 1)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 1.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
         );
@@ -1279,16 +_,16 @@
             LootTable.lootTable()
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(
                             LootItem.lootTableItem(Items.ROTTEN_FLESH)
-                                .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                                .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                                .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                         )
                 )
                 .withPool(
                     LootPool.lootPool()
-                        .setRolls(ContextIntProviders.exactly(1))
+                        .setRolls(ConstantValue.exactly(1.0F))
                         .add(LootItem.lootTableItem(Items.IRON_INGOT))
                         .add(LootItem.lootTableItem(Items.CARROT))
                         .add(LootItem.lootTableItem(Items.POTATO).apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot())))
@@ -1302,38 +_,38 @@
         return LootTable.lootTable()
             .withPool(
                 LootPool.lootPool()
-                    .setRolls(ContextIntProviders.exactly(1))
+                    .setRolls(ConstantValue.exactly(1.0F))
                     .add(
                         LootItem.lootTableItem(Items.PRISMARINE_SHARD)
-                            .apply(SetItemCountFunction.setCount(ContextIntProviders.between(0, 2)))
-                            .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                            .apply(SetItemCountFunction.setCount(UniformGenerator.between(0.0F, 2.0F)))
+                            .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                     )
             )
             .withPool(
                 LootPool.lootPool()
-                    .setRolls(ContextIntProviders.exactly(1))
+                    .setRolls(ConstantValue.exactly(1.0F))
                     .add(
                         LootItem.lootTableItem(Items.COD)
                             .setWeight(3)
-                            .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                            .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                             .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
                     )
                     .add(
                         LootItem.lootTableItem(Items.PRISMARINE_CRYSTALS)
                             .setWeight(2)
-                            .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, ContextFloatProviders.between(0.0F, 1.0F)))
+                            .apply(EnchantedCountIncreaseFunction.lootingMultiplier(this.enchantments, UniformGenerator.between(0.0F, 1.0F)))
                     )
                     .add(EmptyLootItem.emptyItem())
             )
             .withPool(
                 LootPool.lootPool()
-                    .setRolls(ContextIntProviders.exactly(1))
+                    .setRolls(ConstantValue.exactly(1.0F))
                     .add(LootItem.lootTableItem(Blocks.WET_SPONGE))
                     .when(LootItemKilledByPlayerCondition.killedByPlayer())
             )
             .withPool(
                 LootPool.lootPool()
-                    .setRolls(ContextIntProviders.exactly(1))
+                    .setRolls(ConstantValue.exactly(1.0F))
                     .add(
                         NestedLootTable.lootTableReference(this.lootTables.getOrThrow(BuiltInLootTables.FISHING_FISH))
                             .apply(SmeltItemFunction.smelted().when(this.shouldSmeltLoot()))
@@ -1343,7 +_,7 @@
             )
             .withPool(
                 LootPool.lootPool()
-                    .setRolls(ContextIntProviders.exactly(1))
+                    .setRolls(ConstantValue.exactly(1.0F))
                     .add(EmptyLootItem.emptyItem().setWeight(4))
                     .add(LootItem.lootTableItem(Items.TIDE_ARMOR_TRIM_SMITHING_TEMPLATE).setWeight(1))
             );
