--- a/net/minecraft/data/loot/packs/VanillaEquipmentLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaEquipmentLoot.java
@@ -22,7 +_,7 @@
 import net.minecraft.world.level.storage.loot.functions.SetComponentsFunction;
 import net.minecraft.world.level.storage.loot.functions.SetEnchantmentsFunction;
 import net.minecraft.world.level.storage.loot.predicates.LootItemRandomChanceCondition;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
 
 public class VanillaEquipmentLoot implements LootTableSubProvider {
     private final LootTableSubProvider.Context output;
@@ -47,7 +_,7 @@
                 LootTable.lootTable()
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(
                                 NestedLootTable.inlineLootTable(
                                         this.trialChamberEquipment(Items.CHAINMAIL_HELMET, Items.CHAINMAIL_CHESTPLATE, boltTrim).build()
@@ -68,23 +_,23 @@
             .accept(
                 BuiltInLootTables.EQUIPMENT_TRIAL_CHAMBER_MELEE,
                 LootTable.lootTable()
-                    .withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(NestedLootTable.lootTableReference(equipmentTrialChamblerTable)))
+                    .withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(NestedLootTable.lootTableReference(equipmentTrialChamblerTable)))
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.IRON_SWORD).setWeight(4))
                             .add(
                                 LootItem.lootTableItem(Items.IRON_SWORD)
                                     .apply(
                                         new SetEnchantmentsFunction.Builder()
-                                            .withEnchantment(this.enchantments.getOrThrow(Enchantments.SHARPNESS), ContextIntProviders.exactly(1))
+                                            .withEnchantment(this.enchantments.getOrThrow(Enchantments.SHARPNESS), ConstantValue.exactly(1.0F))
                                     )
                             )
                             .add(
                                 LootItem.lootTableItem(Items.IRON_SWORD)
                                     .apply(
                                         new SetEnchantmentsFunction.Builder()
-                                            .withEnchantment(this.enchantments.getOrThrow(Enchantments.KNOCKBACK), ContextIntProviders.exactly(1))
+                                            .withEnchantment(this.enchantments.getOrThrow(Enchantments.KNOCKBACK), ConstantValue.exactly(1.0F))
                                     )
                             )
                             .add(LootItem.lootTableItem(Items.DIAMOND_SWORD))
@@ -94,23 +_,23 @@
             .accept(
                 BuiltInLootTables.EQUIPMENT_TRIAL_CHAMBER_RANGED,
                 LootTable.lootTable()
-                    .withPool(LootPool.lootPool().setRolls(ContextIntProviders.exactly(1)).add(NestedLootTable.lootTableReference(equipmentTrialChamblerTable)))
+                    .withPool(LootPool.lootPool().setRolls(ConstantValue.exactly(1.0F)).add(NestedLootTable.lootTableReference(equipmentTrialChamblerTable)))
                     .withPool(
                         LootPool.lootPool()
-                            .setRolls(ContextIntProviders.exactly(1))
+                            .setRolls(ConstantValue.exactly(1.0F))
                             .add(LootItem.lootTableItem(Items.BOW).setWeight(2))
                             .add(
                                 LootItem.lootTableItem(Items.BOW)
                                     .apply(
                                         new SetEnchantmentsFunction.Builder()
-                                            .withEnchantment(this.enchantments.getOrThrow(Enchantments.POWER), ContextIntProviders.exactly(1))
+                                            .withEnchantment(this.enchantments.getOrThrow(Enchantments.POWER), ConstantValue.exactly(1.0F))
                                     )
                             )
                             .add(
                                 LootItem.lootTableItem(Items.BOW)
                                     .apply(
                                         new SetEnchantmentsFunction.Builder()
-                                            .withEnchantment(this.enchantments.getOrThrow(Enchantments.PUNCH), ContextIntProviders.exactly(1))
+                                            .withEnchantment(this.enchantments.getOrThrow(Enchantments.PUNCH), ConstantValue.exactly(1.0F))
                                     )
                             )
                     )
@@ -121,31 +_,31 @@
         return LootTable.lootTable()
             .withPool(
                 LootPool.lootPool()
-                    .setRolls(ContextIntProviders.exactly(1))
+                    .setRolls(ConstantValue.exactly(1.0F))
                     .when(LootItemRandomChanceCondition.randomChance(0.5F))
                     .add(
                         LootItem.lootTableItem(helmet)
                             .apply(SetComponentsFunction.setComponent(DataComponents.TRIM, trim))
                             .apply(
                                 new SetEnchantmentsFunction.Builder()
-                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.PROTECTION), ContextIntProviders.exactly(4))
-                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.PROJECTILE_PROTECTION), ContextIntProviders.exactly(4))
-                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.FIRE_PROTECTION), ContextIntProviders.exactly(4))
+                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.PROTECTION), ConstantValue.exactly(4.0F))
+                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.PROJECTILE_PROTECTION), ConstantValue.exactly(4.0F))
+                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.FIRE_PROTECTION), ConstantValue.exactly(4.0F))
                             )
                     )
             )
             .withPool(
                 LootPool.lootPool()
-                    .setRolls(ContextIntProviders.exactly(1))
+                    .setRolls(ConstantValue.exactly(1.0F))
                     .when(LootItemRandomChanceCondition.randomChance(0.5F))
                     .add(
                         LootItem.lootTableItem(chestplate)
                             .apply(SetComponentsFunction.setComponent(DataComponents.TRIM, trim))
                             .apply(
                                 new SetEnchantmentsFunction.Builder()
-                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.PROTECTION), ContextIntProviders.exactly(4))
-                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.PROJECTILE_PROTECTION), ContextIntProviders.exactly(4))
-                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.FIRE_PROTECTION), ContextIntProviders.exactly(4))
+                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.PROTECTION), ConstantValue.exactly(4.0F))
+                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.PROJECTILE_PROTECTION), ConstantValue.exactly(4.0F))
+                                    .withEnchantment(this.enchantments.getOrThrow(Enchantments.FIRE_PROTECTION), ConstantValue.exactly(4.0F))
                             )
                     )
             );
