--- a/net/minecraft/data/loot/packs/VanillaChestLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaChestLoot.java
@@ -1135,7 +_,7 @@
                             .add(
                                 LootItem.lootTableItem(Items.STONE_PICKAXE)
                                     .apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1)))
-                                    .apply(SetItemDamageFunction.setDamage(ContextFloatProviders.between(0.15F, 0.8F)))
+                                    /*.apply(SetItemDamageFunction.setDamage(ContextFloatProviders.between(0.15F, 0.8F)))*/ // Paper - Remove set damage lootable from compasses
                                     .setWeight(2)
                             )
                             .add(LootItem.lootTableItem(Items.MILK_BUCKET).apply(SetItemCountFunction.setCount(ContextIntProviders.exactly(1))))
