--- a/net/minecraft/client/renderer/block/dispatch/BlockStateModelPart.java
+++ b/net/minecraft/client/renderer/block/dispatch/BlockStateModelPart.java
@@ -10,10 +_,8 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.extensions.BlockStateModelPartExtension;
-
 @OnlyIn(Dist.CLIENT)
-public interface BlockStateModelPart extends BlockStateModelPartExtension {
+public interface BlockStateModelPart {
     List<BakedQuad> getQuads(@Nullable Direction direction);
 
     boolean useAmbientOcclusion();
