--- a/net/minecraft/client/renderer/block/dispatch/BlockStateModel.java
+++ b/net/minecraft/client/renderer/block/dispatch/BlockStateModel.java
@@ -19,10 +_,8 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
-import net.neoforged.neoforge.client.extensions.BlockStateModelExtension;
-
 @OnlyIn(Dist.CLIENT)
-public interface BlockStateModel extends BlockStateModelExtension {
+public interface BlockStateModel {
     void collectParts(RandomSource random, List<BlockStateModelPart> output);
 
     Material.Baked particleMaterial();
