--- a/net/minecraft/client/renderer/block/dispatch/BlockModelRotation.java
+++ b/net/minecraft/client/renderer/block/dispatch/BlockModelRotation.java
@@ -55,7 +_,7 @@
         return "simple[" + this.orientation.getSerializedName() + "]";
     }
 
-    public record WithUvLock(BlockModelRotation parent) implements ModelState {
+    private record WithUvLock(BlockModelRotation parent) implements ModelState {
         @Override
         public Transformation transformation() {
             return this.parent.transformation;
