--- a/net/minecraft/client/renderer/block/dispatch/ModelState.java
+++ b/net/minecraft/client/renderer/block/dispatch/ModelState.java
@@ -7,10 +_,8 @@
 import org.joml.Matrix4f;
 import org.joml.Matrix4fc;
 
-import net.neoforged.neoforge.client.extensions.ModelStateExtension;
-
 @OnlyIn(Dist.CLIENT)
-public interface ModelState extends ModelStateExtension {
+public interface ModelState {
     Matrix4fc NO_TRANSFORM = new Matrix4f();
 
     default Transformation transformation() {
