--- a/net/minecraft/client/renderer/block/dispatch/WeightedVariants.java
+++ b/net/minecraft/client/renderer/block/dispatch/WeightedVariants.java
@@ -11,10 +_,8 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
-import net.neoforged.neoforge.client.model.DynamicBlockStateModel;
-
 @OnlyIn(Dist.CLIENT)
-public class WeightedVariants implements BlockStateModel, DynamicBlockStateModel {
+public class WeightedVariants implements BlockStateModel {
     private final WeightedList<BlockStateModel> list;
     private final Material.Baked particleMaterial;
     private final @BakedQuad.MaterialFlags int materialFlags;
