--- a/net/minecraft/world/level/storage/loot/providers/number/ints/ResolvableInt.java
+++ b/net/minecraft/world/level/storage/loot/providers/number/ints/ResolvableInt.java
@@ -62,7 +_,7 @@
             return this.getProvider(context).map(provider -> provider.getInt(context)).orElse(defaultValue);
         }
 
-        private Optional<ContextIntProvider> getProvider(LootContext context) {
+        public Optional<ContextIntProvider> getProvider(LootContext context) {
             return context.getResolver().lookupOrThrow(Registries.CONTEXT_INT_PROVIDER).get(this.key).map(Holder.Reference::value);
         }
     }
