--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -24,6 +_,8 @@
 import net.minecraft.world.level.entity.EntityTypeTest;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.slf4j.Logger;
 
 public abstract class BaseSpawner implements net.neoforged.neoforge.common.extensions.IOwnedSpawner {
@@ -49,12 +_,13 @@
         this.getOrCreateNextSpawnData(p_254041_, p_254221_, p_254050_)
             .getEntityToSpawn()
             .putString("id", BuiltInRegistries.ENTITY_TYPE.getKey(p_253682_).toString());
+        this.spawnPotentials = SimpleWeightedRandomList.empty(); // CraftBukkit - SPIGOT-3496, MC-92282
     }
 
     private boolean isNearPlayer(Level p_151344_, BlockPos p_151345_) {
-        return p_151344_.hasNearbyAlivePlayer(
+        return p_151344_.hasNearbyAlivePlayerThatAffectsSpawning(
             (double)p_151345_.getX() + 0.5, (double)p_151345_.getY() + 0.5, (double)p_151345_.getZ() + 0.5, (double)this.requiredPlayerRange
-        );
+        );  // Paper - Affects Spawning API
     }
 
     public void clientTick(Level p_151320_, BlockPos p_151321_) {
@@ -151,19 +_,22 @@
 
                         entity.moveTo(entity.getX(), entity.getY(), entity.getZ(), randomsource.nextFloat() * 360.0F, 0.0F);
                         if (entity instanceof Mob mob) {
-                            if (spawndata.getCustomSpawnRules().isEmpty() && !mob.checkSpawnRules(p_151312_, MobSpawnType.SPAWNER)
-                                || !mob.checkSpawnObstruction(p_151312_)) {
+                            if (!net.neoforged.neoforge.event.EventHooks.checkSpawnPositionSpawner(mob, p_151312_, MobSpawnType.SPAWNER, spawndata, this)) {
                                 continue;
                             }
 
                             boolean flag1 = spawndata.getEntityToSpawn().size() == 1 && spawndata.getEntityToSpawn().contains("id", 8);
-                            if (flag1) {
-                                ((Mob)entity).finalizeSpawn(p_151312_, p_151312_.getCurrentDifficultyAt(entity.blockPosition()), MobSpawnType.SPAWNER, null);
-                            }
+                            // Neo: Patch in FinalizeSpawn for spawners so it may be fired unconditionally, instead of only when vanilla would normally call it.
+                            // The local flag1 is the conditions under which the spawner will normally call Mob#finalizeSpawn.
+                            net.neoforged.neoforge.event.EventHooks.finalizeMobSpawnSpawner(mob, p_151312_, p_151312_.getCurrentDifficultyAt(entity.blockPosition()), MobSpawnType.SPAWNER, null, this, flag1);
 
                             spawndata.getEquipment().ifPresent(mob::equip);
                         }
-
+                        // CraftBukkit start
+                        if (CraftEventFactory.callSpawnerSpawnEvent(entity, p_151313_).isCancelled()) {
+                            continue;
+                        }
+                        entity.spawnReason(CreatureSpawnEvent.SpawnReason.SPAWNER); // Mohist
                         if (!p_151312_.tryAddFreshEntityWithPassengers(entity)) {
                             this.delay(p_151312_, p_151313_);
                             return;
@@ -308,5 +_,12 @@
 
     public double getoSpin() {
         return this.oSpin;
+    }
+
+    @Override
+    @org.jetbrains.annotations.Nullable
+    public com.mojang.datafixers.util.Either<net.minecraft.world.level.block.entity.BlockEntity, Entity> getOwner() {
+        // The vanilla anonymous classes have proper overrides, but we return null here for compatibility.
+        return null;
     }
 }
