--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -24,6 +_,8 @@
 import net.minecraft.world.level.entity.EntityTypeTest;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.slf4j.Logger;
 
 public abstract class BaseSpawner implements net.neoforged.neoforge.common.extensions.IOwnedSpawner {
@@ -44,17 +_,19 @@
     public int maxNearbyEntities = 6;
     public int requiredPlayerRange = 16;
     public int spawnRange = 4;
+    private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
 
     public void setEntityId(EntityType<?> p_253682_, @Nullable Level p_254041_, RandomSource p_254221_, BlockPos p_254050_) {
         this.getOrCreateNextSpawnData(p_254041_, p_254221_, p_254050_)
             .getEntityToSpawn()
             .putString("id", BuiltInRegistries.ENTITY_TYPE.getKey(p_253682_).toString());
+        this.spawnPotentials = SimpleWeightedRandomList.empty(); // CraftBukkit - SPIGOT-3496, MC-92282
     }
 
-    private boolean isNearPlayer(Level p_151344_, BlockPos p_151345_) {
-        return p_151344_.hasNearbyAlivePlayer(
+    public boolean isNearPlayer(Level p_151344_, BlockPos p_151345_) {
+        return p_151344_.hasNearbyAlivePlayerThatAffectsSpawning(
             (double)p_151345_.getX() + 0.5, (double)p_151345_.getY() + 0.5, (double)p_151345_.getZ() + 0.5, (double)this.requiredPlayerRange
-        );
+        );  // Paper - Affects Spawning API
     }
 
     public void clientTick(Level p_151320_, BlockPos p_151321_) {
@@ -77,13 +_,19 @@
     }
 
     public void serverTick(ServerLevel p_151312_, BlockPos p_151313_) {
+        if (spawnCount <= 0 || maxNearbyEntities <= 0) return; // Paper - Ignore impossible spawn tick
+        // Paper start - Configurable mob spawner tick rate
+        if (spawnDelay > 0 && --tickDelay > 0) return;
+        tickDelay = p_151312_.paperConfig().tickRates.mobSpawner;
+        if (tickDelay == -1) { return; } // If disabled
+        // Paper end - Configurable mob spawner tick rate
         if (this.isNearPlayer(p_151312_, p_151313_)) {
-            if (this.spawnDelay == -1) {
+            if (this.spawnDelay == -tickDelay) { // Paper - Configurable mob spawner tick rate
                 this.delay(p_151312_, p_151313_);
             }
 
             if (this.spawnDelay > 0) {
-                this.spawnDelay--;
+                this.spawnDelay -= tickDelay; // Paper - Configurable mob spawner tick rate
             } else {
                 boolean flag = false;
                 RandomSource randomsource = p_151312_.getRandom();
@@ -121,6 +_,23 @@
                             continue;
                         }
 
+                        {
+                            // Paper start - PreCreatureSpawnEvent
+                            com.destroystokyo.paper.event.entity.PreSpawnerSpawnEvent event = new com.destroystokyo.paper.event.entity.PreSpawnerSpawnEvent(
+                                    io.papermc.paper.util.MCUtil.toLocation(p_151312_, d0, d1, d2),
+                                    org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(optional.get()),
+                                    io.papermc.paper.util.MCUtil.toLocation(p_151312_, p_151313_)
+                            );
+                            if (!event.callEvent()) {
+                                flag = true;
+                                if (event.shouldAbortSpawn()) {
+                                    break;
+                                }
+                                continue;
+                            }
+                            // Paper end - PreCreatureSpawnEvent\
+                        }
+
                         Entity entity = EntityType.loadEntityRecursive(compoundtag, p_151312_, p_151310_ -> {
                             p_151310_.moveTo(d0, d1, d2, p_151310_.getYRot(), p_151310_.getXRot());
                             return p_151310_;
@@ -149,21 +_,27 @@
                             return;
                         }
 
+                        entity.preserveMotion = true; // Paper - Fix Entity Teleportation and cancel velocity if teleported; preserve entity motion from tag
                         entity.moveTo(entity.getX(), entity.getY(), entity.getZ(), randomsource.nextFloat() * 360.0F, 0.0F);
                         if (entity instanceof Mob mob) {
-                            if (spawndata.getCustomSpawnRules().isEmpty() && !mob.checkSpawnRules(p_151312_, MobSpawnType.SPAWNER)
-                                || !mob.checkSpawnObstruction(p_151312_)) {
+                            if (!net.neoforged.neoforge.event.EventHooks.checkSpawnPositionSpawner(mob, p_151312_, MobSpawnType.SPAWNER, spawndata, this)) {
                                 continue;
                             }
 
                             boolean flag1 = spawndata.getEntityToSpawn().size() == 1 && spawndata.getEntityToSpawn().contains("id", 8);
-                            if (flag1) {
-                                ((Mob)entity).finalizeSpawn(p_151312_, p_151312_.getCurrentDifficultyAt(entity.blockPosition()), MobSpawnType.SPAWNER, null);
-                            }
+                            // Neo: Patch in FinalizeSpawn for spawners so it may be fired unconditionally, instead of only when vanilla would normally call it.
+                            // The local flag1 is the conditions under which the spawner will normally call Mob#finalizeSpawn.
+                            net.neoforged.neoforge.event.EventHooks.finalizeMobSpawnSpawner(mob, p_151312_, p_151312_.getCurrentDifficultyAt(entity.blockPosition()), MobSpawnType.SPAWNER, null, this, flag1);
 
                             spawndata.getEquipment().ifPresent(mob::equip);
                         }
-
+                        // CraftBukkit start
+                        entity.spawnedViaMobSpawner = true; // Paper
+                        flag = true; // Paper
+                        entity.spawnReason(CreatureSpawnEvent.SpawnReason.SPAWNER); // Mohist  // Paper - Entity#getEntitySpawnReason
+                        if (CraftEventFactory.callSpawnerSpawnEvent(entity, p_151313_).isCancelled()) {
+                            continue;
+                        }
                         if (!p_151312_.tryAddFreshEntityWithPassengers(entity)) {
                             this.delay(p_151312_, p_151313_);
                             return;
@@ -174,8 +_,6 @@
                         if (entity instanceof Mob) {
                             ((Mob)entity).spawnAnim();
                         }
-
-                        flag = true;
                     }
                 }
 
@@ -186,7 +_,7 @@
         }
     }
 
-    private void delay(Level p_151351_, BlockPos p_151352_) {
+    public void delay(Level p_151351_, BlockPos p_151352_) {
         RandomSource randomsource = p_151351_.random;
         if (this.maxSpawnDelay <= this.minSpawnDelay) {
             this.spawnDelay = this.minSpawnDelay;
@@ -199,7 +_,13 @@
     }
 
     public void load(@Nullable Level p_151329_, BlockPos p_151330_, CompoundTag p_151331_) {
-        this.spawnDelay = p_151331_.getShort("Delay");
+        // Paper start - use larger int if set
+        if (p_151331_.contains("Paper.Delay")) {
+            this.spawnDelay = p_151331_.getInt("Paper.Delay");
+        } else {
+            this.spawnDelay = p_151331_.getShort("Delay");
+        }
+        // Paper end
         boolean flag = p_151331_.contains("SpawnData", 10);
         if (flag) {
             SpawnData spawndata = SpawnData.CODEC
@@ -220,9 +_,15 @@
             this.spawnPotentials = SimpleWeightedRandomList.single(this.nextSpawnData != null ? this.nextSpawnData : new SpawnData());
         }
 
-        if (p_151331_.contains("MinSpawnDelay", 99)) {
-            this.minSpawnDelay = p_151331_.getShort("MinSpawnDelay");
-            this.maxSpawnDelay = p_151331_.getShort("MaxSpawnDelay");
+        // Paper start - use ints if set
+        if (p_151331_.contains("Paper.MinSpawnDelay", net.minecraft.nbt.Tag.TAG_ANY_NUMERIC)) {
+            this.minSpawnDelay = p_151331_.getInt("Paper.MinSpawnDelay");
+            this.maxSpawnDelay = p_151331_.getInt("Paper.MaxSpawnDelay");
+            this.spawnCount = p_151331_.getShort("SpawnCount");
+        } else // Paper end
+            if (p_151331_.contains("MinSpawnDelay", 99)) {
+                this.minSpawnDelay = p_151331_.getInt("MinSpawnDelay"); // Paper - short -> int
+                this.maxSpawnDelay = p_151331_.getInt("MaxSpawnDelay"); // Paper - short -> int
             this.spawnCount = p_151331_.getShort("SpawnCount");
         }
 
@@ -239,9 +_,20 @@
     }
 
     public CompoundTag save(CompoundTag p_186382_) {
-        p_186382_.putShort("Delay", (short)this.spawnDelay);
-        p_186382_.putShort("MinSpawnDelay", (short)this.minSpawnDelay);
-        p_186382_.putShort("MaxSpawnDelay", (short)this.maxSpawnDelay);
+        // Paper start
+        if (spawnDelay > Short.MAX_VALUE) {
+            p_186382_.putInt("Paper.Delay", this.spawnDelay);
+        }
+        p_186382_.putShort("Delay", (short) Math.min(Short.MAX_VALUE, this.spawnDelay));
+
+        if (minSpawnDelay > Short.MAX_VALUE || maxSpawnDelay > Short.MAX_VALUE) {
+            p_186382_.putInt("Paper.MinSpawnDelay", this.minSpawnDelay);
+            p_186382_.putInt("Paper.MaxSpawnDelay", this.maxSpawnDelay);
+        }
+
+        p_186382_.putShort("MinSpawnDelay", (short) Math.min(Short.MAX_VALUE, this.minSpawnDelay));
+        p_186382_.putShort("MaxSpawnDelay", (short) Math.min(Short.MAX_VALUE, this.maxSpawnDelay));
+        // Paper end
         p_186382_.putShort("SpawnCount", (short)this.spawnCount);
         p_186382_.putShort("MaxNearbyEntities", (short)this.maxNearbyEntities);
         p_186382_.putShort("RequiredPlayerRange", (short)this.requiredPlayerRange);
@@ -287,7 +_,7 @@
         }
     }
 
-    protected void setNextSpawnData(@Nullable Level p_151325_, BlockPos p_151326_, SpawnData p_151327_) {
+    public void setNextSpawnData(@Nullable Level p_151325_, BlockPos p_151326_, SpawnData p_151327_) {
         this.nextSpawnData = p_151327_;
     }
 
@@ -308,5 +_,12 @@
 
     public double getoSpin() {
         return this.oSpin;
+    }
+
+    @Override
+    @org.jetbrains.annotations.Nullable
+    public com.mojang.datafixers.util.Either<net.minecraft.world.level.block.entity.BlockEntity, Entity> getOwner() {
+        // The vanilla anonymous classes have proper overrides, but we return null here for compatibility.
+        return null;
     }
 }
