--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -60,9 +_,11 @@
     public int maxNearbyEntities = 6;
     public int requiredPlayerRange = 16;
     public int spawnRange = 4;
+    private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
 
     public void setEntityId(EntityType<?> type, @Nullable Level level, RandomSource random, BlockPos pos) {
         this.getOrCreateNextSpawnData(level, random, pos).getEntityToSpawn().putString("id", BuiltInRegistries.ENTITY_TYPE.getKey(type).toString());
+        this.spawnPotentials = WeightedList.of(); // CraftBukkit - SPIGOT-3496, MC-92282
     }
 
     public void setEntityData(TypedEntityData<EntityType<?>> entityData, @Nullable Level level, RandomSource random, BlockPos pos) {
@@ -71,7 +_,8 @@
     }
 
     public boolean isNearPlayer(Level level, BlockPos pos) {
-        return level.hasNearbyAlivePlayer(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange);
+        if (level.purpurConfig.spawnerDeactivateByRedstone && level.hasNeighborSignal(pos)) return false; // Purpur - Redstone deactivates spawners
+        return level.hasNearbyAlivePlayerThatAffectsSpawning(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange); // Paper - Affects Spawning API
     }
 
     public void clientTick(Level level, BlockPos pos) {
@@ -94,13 +_,19 @@
     }
 
     public void serverTick(ServerLevel level, BlockPos pos) {
+        if (spawnCount <= 0 || maxNearbyEntities <= 0) return; // Paper - Ignore impossible spawn tick
+        // Paper start - Configurable mob spawner tick rate
+        if (spawnDelay > 0 && --tickDelay > 0) return;
+        tickDelay = level.paperConfig().tickRates.mobSpawner;
+        if (tickDelay == -1) { return; } // If disabled
+        // Paper end - Configurable mob spawner tick rate
         if (this.isNearPlayer(level, pos) && level.isSpawnerBlockEnabled()) {
-            if (this.spawnDelay == -1) {
+            if (this.spawnDelay == -tickDelay) { // Paper - Configurable mob spawner tick rate
                 this.delay(level, pos);
             }
 
             if (this.spawnDelay > 0) {
-                this.spawnDelay--;
+                this.spawnDelay -= tickDelay; // Paper - Configurable mob spawner tick rate
             } else {
                 boolean delay = false;
                 RandomSource random = level.getRandom();
@@ -138,6 +_,21 @@
                                 continue;
                             }
 
+                            // Paper start - PreCreatureSpawnEvent
+                            com.destroystokyo.paper.event.entity.PreSpawnerSpawnEvent event = new com.destroystokyo.paper.event.entity.PreSpawnerSpawnEvent(
+                                    org.bukkit.craftbukkit.util.CraftLocation.toBukkit(spawnPos, level),
+                                    org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(entityType.get()),
+                                    org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level)
+                            );
+                            if (!event.callEvent()) {
+                                delay = true;
+                                if (event.shouldAbortSpawn()) {
+                                    break;
+                                }
+                                continue;
+                            }
+                            // Paper end - PreCreatureSpawnEvent
+
                             Entity entity = EntityType.loadEntityRecursive(input, level, EntitySpawnReason.SPAWNER, e -> {
                                 e.snapTo(spawnPos.x, spawnPos.y, spawnPos.z, e.getYRot(), e.getXRot());
                                 return e;
@@ -160,21 +_,33 @@
 
                             entity.snapTo(entity.getX(), entity.getY(), entity.getZ(), random.nextFloat() * 360.0F, 0.0F);
                             if (entity instanceof Mob mob) {
-                                if (nextSpawnData.getCustomSpawnRules().isEmpty() && !mob.checkSpawnRules(level, EntitySpawnReason.SPAWNER)
-                                    || !mob.checkSpawnObstruction(level)) {
+                                if (!net.neoforged.neoforge.event.EventHooks.checkSpawnPositionSpawner(mob, level, EntitySpawnReason.SPAWNER, nextSpawnData, this)) {
                                     continue;
                                 }
 
                                 boolean hasNoConfiguration = nextSpawnData.getEntityToSpawn().size() == 1
                                     && nextSpawnData.getEntityToSpawn().getString("id").isPresent();
-                                if (hasNoConfiguration) {
-                                    ((Mob)entity).finalizeSpawn(level, level.getCurrentDifficultyAt(entity.blockPosition()), EntitySpawnReason.SPAWNER, null);
-                                }
+                                // Neo: Patch in FinalizeSpawn for spawners so it may be fired unconditionally, instead of only when vanilla would normally call it.
+                                // The local hasNoConfiguration is the conditions under which the spawner will normally call Mob#finalizeSpawn.
+                                net.neoforged.neoforge.event.EventHooks.finalizeMobSpawnSpawner(mob, level, level.getCurrentDifficultyAt(entity.blockPosition()), EntitySpawnReason.SPAWNER, null, this, hasNoConfiguration);
 
                                 nextSpawnData.getEquipment().ifPresent(mob::equip);
+                                // Spigot start
+                                if (mob.level().spigotConfig.nerfSpawnerMobs) {
+                                    mob.aware = false;
+                                }
+                                // Spigot end
                             }
 
-                            if (!level.tryAddFreshEntityWithPassengers(entity)) {
+                            // Paper start
+                            entity.spawnedViaMobSpawner = true;
+                            entity.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER;
+                            delay = true;
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callSpawnerSpawnEvent(entity, pos).isCancelled()) {
+                                continue;
+                            }
+                            if (!level.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER)) {
+                                // Paper end
                                 this.delay(level, pos);
                                 return;
                             }
@@ -185,7 +_,7 @@
                                 mob.spawnAnim();
                             }
 
-                            delay = true;
+                            // delay = true; // Paper - moved up above cancellable event
                         }
                     }
                 }
@@ -212,12 +_,14 @@
     }
 
     public void load(@Nullable Level level, BlockPos pos, ValueInput input) {
-        this.spawnDelay = input.getShortOr("Delay", (short)20);
+        this.spawnDelay = input.getIntOr("Paper.Delay", input.getShortOr("Delay", (short) 20)); // Paper - use int if set
         input.read("SpawnData", SpawnData.CODEC).ifPresent(nextSpawnData -> this.setNextSpawnData(level, pos, nextSpawnData));
         this.spawnPotentials = input.read("SpawnPotentials", SpawnData.LIST_CODEC)
             .orElseGet(() -> WeightedList.of(this.nextSpawnData != null ? this.nextSpawnData : new SpawnData()));
-        this.minSpawnDelay = input.getIntOr("MinSpawnDelay", 200);
-        this.maxSpawnDelay = input.getIntOr("MaxSpawnDelay", 800);
+        // Paper start - use int if set
+        this.minSpawnDelay = input.getIntOr("Paper.MinSpawnDelay", input.getIntOr("MinSpawnDelay", 200));
+        this.maxSpawnDelay = input.getIntOr("Paper.MaxSpawnDelay", input.getIntOr("MaxSpawnDelay", 800));
+        // Paper end - use int if set
         this.spawnCount = input.getIntOr("SpawnCount", 4);
         this.maxNearbyEntities = input.getIntOr("MaxNearbyEntities", 6);
         this.requiredPlayerRange = input.getIntOr("RequiredPlayerRange", 16);
@@ -226,9 +_,19 @@
     }
 
     public void save(ValueOutput output) {
-        output.putShort("Delay", (short)this.spawnDelay);
-        output.putShort("MinSpawnDelay", (short)this.minSpawnDelay);
-        output.putShort("MaxSpawnDelay", (short)this.maxSpawnDelay);
+        // Paper start
+        if (this.spawnDelay > Short.MAX_VALUE) {
+            output.putInt("Paper.Delay", this.spawnDelay);
+        }
+        output.putShort("Delay", (short) Math.min(Short.MAX_VALUE, this.spawnDelay));
+
+        if (this.minSpawnDelay > Short.MAX_VALUE || this.maxSpawnDelay > Short.MAX_VALUE) {
+            output.putInt("Paper.MinSpawnDelay", this.minSpawnDelay);
+            output.putInt("Paper.MaxSpawnDelay", this.maxSpawnDelay);
+        }
+        output.putShort("MinSpawnDelay", (short) Math.min(Short.MAX_VALUE, this.minSpawnDelay));
+        output.putShort("MaxSpawnDelay", (short) Math.min(Short.MAX_VALUE, this.maxSpawnDelay));
+        // Paper end
         output.putShort("SpawnCount", (short)this.spawnCount);
         output.putShort("MaxNearbyEntities", (short)this.maxNearbyEntities);
         output.putShort("RequiredPlayerRange", (short)this.requiredPlayerRange);
@@ -287,5 +_,11 @@
 
     public double getOSpin() {
         return this.oSpin;
+    }
+
+    @Override
+    public com.mojang.datafixers.util.@Nullable Either<net.minecraft.world.level.block.entity.BlockEntity, Entity> getOwner() {
+        // The vanilla anonymous classes have proper overrides, but we return null here for compatibility.
+        return null;
     }
 }
