--- a/net/minecraft/world/level/Explosion.java
+++ b/net/minecraft/world/level/Explosion.java
@@ -1,7 +_,9 @@
 package net.minecraft.world.level;
 
+import com.mohistmc.youer.YouerConfig;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
+import com.mohistmc.youer.plugins.ban.bans.BanEvents;
 import com.mojang.datafixers.util.Pair;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import java.util.ArrayList;
@@ -38,6 +_,9 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class Explosion {
     private static final ExplosionDamageCalculator EXPLOSION_DAMAGE_CALCULATOR = new ExplosionDamageCalculator();
@@ -59,6 +_,10 @@
     private final Holder<SoundEvent> explosionSound;
     private final ObjectArrayList<BlockPos> toBlow = new ObjectArrayList<>();
     private final Map<Player, Vec3> hitPlayers = Maps.newHashMap();
+    // CraftBukkit - add field
+    public boolean wasCanceled = BanEvents.banExplosion(); // CraftBukkit - add field
+    public float yield;
+    // CraftBukkit end
 
     public static DamageSource getDefaultDamageSource(Level p_312716_, @Nullable Entity p_312608_) {
         return p_312716_.damageSources().explosion(p_312608_, getIndirectSourceEntityInternal(p_312608_));
@@ -154,7 +_,7 @@
     ) {
         this.level = p_46024_;
         this.source = p_46025_;
-        this.radius = p_46029_;
+        this.radius = (float) Math.max(p_46029_, 0.0); // CraftBukkit - clamp bad values
         this.x = p_46026_;
         this.y = p_46027_;
         this.z = p_46028_;
@@ -165,6 +_,7 @@
         this.smallExplosionParticles = p_312158_;
         this.largeExplosionParticles = p_311904_;
         this.explosionSound = p_320270_;
+        this.yield = this.blockInteraction == Explosion.BlockInteraction.DESTROY_WITH_DECAY ? 1.0F / this.radius : 1.0F; // CraftBukkit
     }
 
     private ExplosionDamageCalculator makeDamageCalculator(@Nullable Entity p_46063_) {
@@ -214,6 +_,11 @@
     }
 
     public void explode() {
+        // CraftBukkit start
+        if (this.radius < 0.1F) {
+            return;
+        }
+        // CraftBukkit end
         this.level.gameEvent(this.source, GameEvent.EXPLODE, new Vec3(this.x, this.y, this.z));
         Set<BlockPos> set = Sets.newHashSet();
         int i = 16;
@@ -268,7 +_,8 @@
         int i1 = Mth.floor(this.y + (double)f2 + 1.0);
         int j2 = Mth.floor(this.z - (double)f2 - 1.0);
         int j1 = Mth.floor(this.z + (double)f2 + 1.0);
-        List<Entity> list = this.level.getEntities(this.source, new AABB((double)k1, (double)i2, (double)j2, (double)l1, (double)i1, (double)j1));
+        List<Entity> list = this.level.getEntities(this.source, new AABB((double)k1, (double)i2, (double)j2, (double)l1, (double)i1, (double)j1), (com.google.common.base.Predicate<Entity>) entity -> entity.isAlive() && !entity.isSpectator()); // Paper - Fix lag from explosions processing dead entities
+        net.neoforged.neoforge.event.EventHooks.onExplosionDetonate(this.level, this, list, f2);
         Vec3 vec3 = new Vec3(this.x, this.y, this.z);
 
         for (Entity entity : list) {
@@ -284,13 +_,40 @@
                         d7 /= d12;
                         d9 /= d12;
                         if (this.damageCalculator.shouldDamageEntity(this, entity)) {
-                            entity.hurt(this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity));
+                            // CraftBukkit start
+                            // Special case ender dragon only give knockback if no damage is cancelled
+                            // Thinks to note:
+                            // - Setting a velocity to a ComplexEntityPart is ignored (and therefore not needed)
+                            // - Damaging ComplexEntityPart while forward the damage to EntityEnderDragon
+                            // - Damaging EntityEnderDragon does nothing
+                            // - EntityEnderDragon hitbock always covers the other parts and is therefore always present
+                            if (entity instanceof net.minecraft.world.entity.boss.EnderDragonPart) {
+                                continue;
+                            }
+
+                            entity.lastDamageCancelled = false;
+
+                            if (entity instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragon) {
+                                for (net.minecraft.world.entity.boss.EnderDragonPart entityComplexPart : ((net.minecraft.world.entity.boss.enderdragon.EnderDragon) entity).subEntities) {
+                                    // Calculate damage separately for each EntityComplexPart
+                                    if (list.contains(entityComplexPart)) {
+                                        entityComplexPart.hurt(this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity));
+                                    }
+                                }
+                            } else {
+                                entity.hurt(this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity));
+                            }
+
+                            if (entity.lastDamageCancelled) { // SPIGOT-5339, SPIGOT-6252, SPIGOT-6777: Skip entity if damage event was cancelled
+                                continue;
+                            }
+                            // CraftBukkit end
                         }
 
                         double d13 = (1.0 - d11) * (double)getSeenPercent(vec3, entity) * (double)this.damageCalculator.getKnockbackMultiplier(entity);
                         double d10;
                         if (entity instanceof LivingEntity livingentity) {
-                            d10 = d13 * (1.0 - livingentity.getAttributeValue(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE));
+                            d10 = entity instanceof Player && this.level.paperConfig().environment.disableExplosionKnockback ? 0 : d13 * (1.0 - livingentity.getAttributeValue(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE));
                         } else {
                             d10 = d13;
                         }
@@ -299,10 +_,21 @@
                         d7 *= d10;
                         d9 *= d10;
                         Vec3 vec31 = new Vec3(d5, d7, d9);
+
+                        // CraftBukkit start - Call EntityKnockbackEvent
+                        if (entity instanceof LivingEntity) {
+                            // Paper start - knockback events
+                            io.papermc.paper.event.entity.EntityKnockbackEvent event = CraftEventFactory.callEntityKnockbackEvent((org.bukkit.craftbukkit.entity.CraftLivingEntity) entity.getBukkitEntity(), this.source, this.damageSource.getEntity() != null ? this.damageSource.getEntity() : this.source, io.papermc.paper.event.entity.EntityKnockbackEvent.Cause.EXPLOSION, d13, vec31);
+                            vec31 = event.isCancelled() ? Vec3.ZERO : org.bukkit.craftbukkit.util.CraftVector.toNMS(event.getKnockback());
+                            // Paper end - knockback events
+                        }
+                        // CraftBukkit end
+
+                        vec31 = net.neoforged.neoforge.event.EventHooks.getExplosionKnockback(this.level, this, entity, vec31);
                         entity.setDeltaMovement(entity.getDeltaMovement().add(vec31));
                         if (entity instanceof Player) {
                             Player player = (Player)entity;
-                            if (!player.isSpectator() && (!player.isCreative() || !player.getAbilities().flying)) {
+                            if (!player.isSpectator() && (!player.isCreative() || !player.getAbilities().flying) && !level.paperConfig().environment.disableExplosionKnockback) { // Paper - Option to disable explosion knockback
                                 this.hitPlayers.put(player, vec31);
                             }
                         }
@@ -314,7 +_,14 @@
         }
     }
 
+    // Mohist start
+    org.bukkit.World bworld;
+    Location location;
+    List<org.bukkit.block.Block> blockList;
+    List<org.bukkit.block.Block> bukkitBlocks;
+
     public void finalizeExplosion(boolean p_46076_) {
+        if (BanEvents.banExplosion()) return;
         if (this.level.isClientSide) {
             this.level
                 .playLocalSound(
@@ -344,9 +_,61 @@
         if (flag) {
             this.level.getProfiler().push("explosion_blocks");
             List<Pair<ItemStack, BlockPos>> list = new ArrayList<>();
+
+            // CraftBukkit start
+            bworld = this.level.getWorld();
+            location = new org.bukkit.Location(bworld, this.x, this.y, this.z);
+
+            blockList = new ObjectArrayList<>();
+            for (int i1 = this.toBlow.size() - 1; i1 >= 0; i1--) {
+                BlockPos cpos = this.toBlow.get(i1);
+                org.bukkit.block.Block bblock = bworld.getBlockAt(cpos.getX(), cpos.getY(), cpos.getZ());
+                if (!bblock.getType().isAir()) {
+                    blockList.add(bblock);
+                }
+            }
+
+            if (this.source != null) {
+                org.bukkit.event.entity.EntityExplodeEvent event = CraftEventFactory.callEntityExplodeEvent(this.source, blockList, this.yield, getBlockInteraction());
+                this.wasCanceled = event.isCancelled();
+                bukkitBlocks = event.blockList();
+                this.yield = event.getYield();
+            } else {
+                org.bukkit.block.Block block = location.getBlock();
+                org.bukkit.block.BlockState blockState = (damageSource.getDirectBlockState() != null) ? damageSource.getDirectBlockState() : block.getState();
+                org.bukkit.event.block.BlockExplodeEvent event = CraftEventFactory.callBlockExplodeEvent(block, blockState, blockList, this.yield, getBlockInteraction());
+                this.wasCanceled = event.isCancelled();
+                bukkitBlocks = event.blockList();
+                this.yield = event.getYield();
+            }
+
+            this.toBlow.clear();
+
+            for (org.bukkit.block.Block bblock : bukkitBlocks) {
+                BlockPos coords = new BlockPos(bblock.getX(), bblock.getY(), bblock.getZ());
+                toBlow.add(coords);
+            }
+
+            if (this.wasCanceled) {
+                return;
+            }
+            // CraftBukkit end
             Util.shuffle(this.toBlow, this.level.random);
 
             for (BlockPos blockpos : this.toBlow) {
+                // CraftBukkit start - TNTPrimeEvent
+                BlockState blockstate = this.level.getBlockState(blockpos);
+                Block block = blockstate.getBlock();
+                if (block instanceof net.minecraft.world.level.block.TntBlock) {
+                    Entity sourceEntity = source == null ? null : source;
+                    BlockPos sourceBlock = sourceEntity == null ? BlockPos.containing(this.x, this.y, this.z) : null;
+                    if (!CraftEventFactory.callTNTPrimeEvent(this.level, blockpos, org.bukkit.event.block.TNTPrimeEvent.PrimeCause.EXPLOSION, sourceEntity, sourceBlock)) {
+                        this.level.sendBlockUpdated(blockpos, net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(), blockstate, 3); // Update the block on the client
+                        continue;
+                    }
+                }
+                // CraftBukkit end
+
                 this.level
                     .getBlockState(blockpos)
                     .onExplosionHit(this.level, blockpos, this, (p_311741_, p_311742_) -> addOrAppendStack(list, p_311741_, p_311742_));
@@ -362,15 +_,20 @@
         if (this.fire) {
             for (BlockPos blockpos1 : this.toBlow) {
                 if (this.random.nextInt(3) == 0
-                    && this.level.getBlockState(blockpos1).isAir()
-                    && this.level.getBlockState(blockpos1.below()).isSolidRender(this.level, blockpos1.below())) {
-                    this.level.setBlockAndUpdate(blockpos1, BaseFireBlock.getState(this.level, blockpos1));
+                        && this.level.getBlockState(blockpos1).isAir()
+                        && this.level.getBlockState(blockpos1.below()).isSolidRender(this.level, blockpos1.below())) {
+                    // CraftBukkit start - Ignition by explosion
+                    if (!CraftEventFactory.callBlockIgniteEvent(this.level, blockpos1, this).isCancelled()) {
+                        this.level.setBlockAndUpdate(blockpos1, BaseFireBlock.getState(this.level, blockpos1));
+                    }
+                    // CraftBukkit end
                 }
             }
         }
     }
 
     private static void addOrAppendStack(List<Pair<ItemStack, BlockPos>> p_312455_, ItemStack p_312913_, BlockPos p_312738_) {
+        if (p_312913_.isEmpty()) return; // CraftBukkit - SPIGOT-5425
         for (int i = 0; i < p_312455_.size(); i++) {
             Pair<ItemStack, BlockPos> pair = p_312455_.get(i);
             ItemStack itemstack = pair.getFirst();
