--- a/net/minecraft/world/level/GameRules.java
+++ b/net/minecraft/world/level/GameRules.java
@@ -543,13 +_,13 @@
             this.onChanged(p_46371_.getSource().getServer());
         }
 
-        public void onChanged(@Nullable MinecraftServer p_46369_) {
+        protected void onChanged(@Nullable MinecraftServer p_46369_) {
             if (p_46369_ != null) {
                 this.type.callback.accept(p_46369_, this.getSelf());
             }
         }
 
-        public abstract void deserialize(String p_46367_);
+        protected abstract void deserialize(String p_46367_);
 
         public abstract String serialize();
 
