--- a/net/minecraft/world/level/GameRules.java
+++ b/net/minecraft/world/level/GameRules.java
@@ -407,7 +_,7 @@
         }
 
         @Override
-        protected void deserialize(String p_46234_) {
+        public void deserialize(String p_46234_) {
             this.value = Boolean.parseBoolean(p_46234_);
         }
 
@@ -517,7 +_,7 @@
         }
 
         @Override
-        protected void deserialize(String p_46307_) {
+        public void deserialize(String p_46307_) {
             this.value = safeParse(p_46307_);
         }
 
