--- a/net/minecraft/world/level/LevelReader.java
+++ b/net/minecraft/world/level/LevelReader.java
@@ -16,9 +_,6 @@
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeManager;
 import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.level.blockscan.BlockMatcher;
-import net.minecraft.world.level.blockscan.BoxBlockMatcher;
-import net.minecraft.world.level.blockscan.OrderedBlockMatcher;
 import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.level.dimension.DimensionType;
@@ -28,9 +_,22 @@
 
 import net.neoforged.neoforge.common.extensions.ILevelReaderExtension;
 
-public interface LevelReader extends BlockAndLightGetter, CollisionGetter, SignalGetter, BiomeManager.NoiseBiomeSource, ILevelReaderExtension {
+public interface LevelReader extends BlockAndLightGetter, CollisionGetter, SignalGetter, BiomeManager.NoiseBiomeSource, ILevelReaderExtension, ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevelReader { // Paper - rewrite chunk system
+    // Paper start - rewrite chunk system
+    @Override
+    public default ChunkAccess moonrise$syncLoadNonFull(final int chunkX, final int chunkZ, final ChunkStatus status) {
+        if (status == null || status.isOrAfter(ChunkStatus.FULL)) {
+            throw new IllegalArgumentException("Status: " + status.toString());
+        }
+        return ((LevelReader)this).getChunk(chunkX, chunkZ, status, true);
+    }
+    // Paper end - rewrite chunk system
+
     @Nullable ChunkAccess getChunk(final int chunkX, final int chunkZ, final ChunkStatus targetStatus, final boolean loadOrGenerate);
 
+    @Nullable ChunkAccess getChunkIfLoadedImmediately(int x, int z); // Paper - ifLoaded api (we need this since current impl blocks if the chunk is loading)
+    default @Nullable ChunkAccess getChunkIfLoadedImmediately(BlockPos pos) { return this.getChunkIfLoadedImmediately(pos.getX() >> 4, pos.getZ() >> 4); }
+
     @Deprecated
     boolean hasChunk(int chunkX, int chunkZ);
 
@@ -143,7 +_,26 @@
     }
 
     default boolean containsAnyLiquid(AABB box) {
-        return this.findBlocksIn(box).filterState(state -> !state.getFluidState().isEmpty()).anyMatched();
+        int x0 = Mth.floor(box.minX);
+        int x1 = Mth.ceil(box.maxX);
+        int y0 = Mth.floor(box.minY);
+        int y1 = Mth.ceil(box.maxY);
+        int z0 = Mth.floor(box.minZ);
+        int z1 = Mth.ceil(box.maxZ);
+        BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
+
+        for (int x = x0; x < x1; x++) {
+            for (int y = y0; y < y1; y++) {
+                for (int z = z0; z < z1; z++) {
+                    BlockState blockState = this.getBlockState(pos.set(x, y, z));
+                    if (!blockState.getFluidState().isEmpty()) {
+                        return true;
+                    }
+                }
+            }
+        }
+
+        return false;
     }
 
     default int getMaxLocalRawBrightness(BlockPos pos) {
@@ -158,30 +_,6 @@
 
     default int getEffectiveSkyBrightness(BlockPos pos) {
         return this.getBrightness(LightLayer.SKY, pos) - this.getSkyDarken();
-    }
-
-    default BlockMatcher findBlocksIn(AABB box) {
-        return new BoxBlockMatcher(this, box);
-    }
-
-    default BlockMatcher findBlocksIn(BlockPos from, BlockPos to) {
-        return new BoxBlockMatcher(this, from, to);
-    }
-
-    default OrderedBlockMatcher findBlocksIn(Iterable<BlockPos> positions) {
-        return new OrderedBlockMatcher(this, positions);
-    }
-
-    default OrderedBlockMatcher findBlocksInBoxByManhattanDistance(BlockPos origin, int range) {
-        return this.findBlocksIn(BlockPos.withinBoxByManhattanDistance(origin, range, range, range));
-    }
-
-    default OrderedBlockMatcher findBlocksInBoxByManhattanDistance(BlockPos origin, int rangeXZ, int rangeY) {
-        return this.findBlocksIn(BlockPos.withinBoxByManhattanDistance(origin, rangeXZ, rangeY, rangeXZ));
-    }
-
-    default OrderedBlockMatcher findBlocksInManhattan(BlockPos origin, int range) {
-        return this.findBlocksIn(BlockPos.withinManhattan(origin, range));
     }
 
     @Deprecated
