--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -6,6 +_,7 @@
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
@@ -40,18 +_,23 @@
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.levelgen.structure.BuiltinStructures;
 import net.minecraft.world.level.levelgen.structure.Structure;
-import net.minecraft.world.level.levelgen.structure.structures.NetherFortressStructure;
 import net.minecraft.world.level.material.FluidState;
+import net.minecraft.world.level.storage.LevelData;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.util.CraftSpawnCategory;
+import org.bukkit.entity.SpawnCategory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.slf4j.Logger;
+import org.spigotmc.AsyncCatcher;
 
+// TODO: ForgeHooks.canEntitySpawn
 public final class NaturalSpawner {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int MIN_SPAWN_DISTANCE = 24;
     public static final int SPAWN_DISTANCE_CHUNK = 8;
     public static final int SPAWN_DISTANCE_BLOCK = 128;
     static final int MAGIC_NUMBER = (int)Math.pow(17.0, 2.0);
-    private static final MobCategory[] SPAWNING_CATEGORIES = Stream.of(MobCategory.values())
+    public static final MobCategory[] SPAWNING_CATEGORIES = Stream.of(MobCategory.values())
         .filter(p_47037_ -> p_47037_ != MobCategory.MISC)
         .toArray(MobCategory[]::new);
 
@@ -69,8 +_,15 @@
                 continue;
             }
 
-            MobCategory mobcategory = entity.getType().getCategory();
+            MobCategory mobcategory = entity.getClassification(true);
             if (mobcategory != MobCategory.MISC) {
+                // Paper start - Only count natural spawns
+                if (entity.level() instanceof ServerLevel serverLevel && !serverLevel.paperConfig().entities.spawning.countAllMobsForSpawning &&
+                        !(entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL ||
+                                entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN)) {
+                    continue;
+                }
+                // Paper end - Only count natural spawns
                 BlockPos blockpos = entity.blockPosition();
                 p_186527_.query(
                     ChunkPos.asLong(blockpos),
@@ -103,8 +_,21 @@
         ServerLevel p_47030_, LevelChunk p_47031_, NaturalSpawner.SpawnState p_47032_, boolean p_47033_, boolean p_47034_, boolean p_47035_
     ) {
         p_47030_.getProfiler().push("spawner");
+        LevelData worlddata = p_47030_.getLevelData(); // CraftBukkit - Other mob type spawn tick rate
 
         for (MobCategory mobcategory : SPAWNING_CATEGORIES) {
+            // CraftBukkit start - Use per-world spawn limits
+            boolean spawnThisTick = true;
+            int limit = mobcategory.getMaxInstancesPerChunk();
+            SpawnCategory spawnCategory = CraftSpawnCategory.toBukkit(mobcategory);
+            if (CraftSpawnCategory.isValidForLimits(spawnCategory)) {
+                spawnThisTick = p_47030_.ticksPerSpawnCategory.getLong(spawnCategory) != 0 && worlddata.getGameTime() % p_47030_.ticksPerSpawnCategory.getLong(spawnCategory) == 0;
+                limit = p_47030_.getWorld().getSpawnLimit(spawnCategory);
+            }
+            if (!spawnThisTick || limit == 0) {
+                continue;
+            }
+            p_47032_.canSpawnForCategory$limit = limit;
             if ((p_47033_ || !mobcategory.isFriendly())
                 && (p_47034_ || mobcategory.isFriendly())
                 && (p_47035_ || !mobcategory.isPersistent())
@@ -116,6 +_,17 @@
         p_47030_.getProfiler().pop();
     }
 
+    // Paper start - Add mobcaps commands
+    public static int globalLimitForCategory(final ServerLevel level, final MobCategory category, final int spawnableChunkCount) {
+        final int categoryLimit = level.getWorld().getSpawnLimitUnsafe(CraftSpawnCategory.toBukkit(category));
+        if (categoryLimit < 1) {
+            return categoryLimit;
+        }
+        return categoryLimit * spawnableChunkCount / NaturalSpawner.MAGIC_NUMBER;
+    }
+    // Paper end - Add mobcaps commands
+
+
     public static void spawnCategoryForChunk(
         MobCategory p_47046_, ServerLevel p_47047_, LevelChunk p_47048_, NaturalSpawner.SpawnPredicate p_47049_, NaturalSpawner.AfterSpawnCallback p_47050_
     ) {
@@ -181,10 +_,13 @@
                                     + p_47040_.random.nextInt(1 + mobspawnsettings$spawnerdata.maxCount - mobspawnsettings$spawnerdata.minCount);
                             }
 
-                            if (isValidSpawnPostitionForType(
-                                    p_47040_, p_47039_, structuremanager, chunkgenerator, mobspawnsettings$spawnerdata, blockpos$mutableblockpos, d2
-                                )
-                                && p_47043_.test(mobspawnsettings$spawnerdata.type, blockpos$mutableblockpos, p_47041_)) {
+                            // Paper start - PreCreatureSpawnEvent
+                            isValidSpawnPostitionForType(p_47040_, p_47039_, structuremanager, chunkgenerator, mobspawnsettings$spawnerdata, blockpos$mutableblockpos, d2);
+                            PreSpawnStatus doSpawning = isValidSpawnPostitionForType$PreSpawnStatus.getAndSet(PreSpawnStatus.FAIL);
+                            if (doSpawning == PreSpawnStatus.ABORT) {
+                                return;
+                            }
+                            if (doSpawning == PreSpawnStatus.SUCCESS && p_47043_.test(mobspawnsettings$spawnerdata.type, blockpos$mutableblockpos, p_47041_)) {
                                 Mob mob = getMobForSpawn(p_47040_, mobspawnsettings$spawnerdata.type);
                                 if (mob == null) {
                                     return;
@@ -197,9 +_,10 @@
                                     );
                                     j++;
                                     l1++;
+                                    mob.spawnReason((mob instanceof net.minecraft.world.entity.animal.Ocelot && !((org.bukkit.entity.Ageable) mob.getBukkitEntity()).isAdult()) ? CreatureSpawnEvent.SpawnReason.OCELOT_BABY : CreatureSpawnEvent.SpawnReason.NATURAL);
                                     p_47040_.addFreshEntityWithPassengers(mob);
                                     p_47044_.run(mob, p_47041_);
-                                    if (j >= mob.getMaxSpawnClusterSize()) {
+                                    if (j >= net.neoforged.neoforge.event.EventHooks.getMaxSpawnClusterSize(mob)) {
                                         return;
                                     }
 
@@ -226,6 +_,13 @@
         }
     }
 
+    private enum PreSpawnStatus {
+        FAIL,
+        SUCCESS,
+        CANCELLED,
+        ABORT
+    }
+    private static final AtomicReference<PreSpawnStatus> isValidSpawnPostitionForType$PreSpawnStatus = new AtomicReference<>(PreSpawnStatus.FAIL);
     private static boolean isValidSpawnPostitionForType(
         ServerLevel p_220422_,
         MobCategory p_220423_,
@@ -235,20 +_,52 @@
         BlockPos.MutableBlockPos p_220427_,
         double p_220428_
     ) {
+        // Paper end - PreCreatureSpawnEvent
         EntityType<?> entitytype = p_220426_.type;
+
+        if (!AsyncCatcher.catchAsync()) {
+            // Paper start - PreCreatureSpawnEvent
+            com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
+                    io.papermc.paper.util.MCUtil.toLocation(p_220422_, p_220427_),
+                    org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(entitytype), CreatureSpawnEvent.SpawnReason.NATURAL
+            );
+            if (!event.callEvent()) {
+                if (event.shouldAbortSpawn()) {
+                    isValidSpawnPostitionForType$PreSpawnStatus.set(PreSpawnStatus.ABORT);
+                    return false;
+                }
+                isValidSpawnPostitionForType$PreSpawnStatus.set(PreSpawnStatus.CANCELLED);
+                return false;
+            }
+            // Paper end - PreCreatureSpawnEvent
+        }
         if (entitytype.getCategory() == MobCategory.MISC) {
+            isValidSpawnPostitionForType$PreSpawnStatus.set(PreSpawnStatus.FAIL);
             return false;
         } else if (!entitytype.canSpawnFarFromPlayer()
-            && p_220428_ > (double)(entitytype.getCategory().getDespawnDistance() * entitytype.getCategory().getDespawnDistance())) {
+                && p_220428_ > (double) (entitytype.getCategory().getDespawnDistance() * entitytype.getCategory().getDespawnDistance())) {
+            isValidSpawnPostitionForType$PreSpawnStatus.set(PreSpawnStatus.FAIL);
             return false;
         } else if (!entitytype.canSummon() || !canSpawnMobAt(p_220422_, p_220424_, p_220425_, p_220423_, p_220426_, p_220427_)) {
+            isValidSpawnPostitionForType$PreSpawnStatus.set(PreSpawnStatus.FAIL);
             return false;
         } else if (!SpawnPlacements.isSpawnPositionOk(entitytype, p_220422_, p_220427_)) {
+            isValidSpawnPostitionForType$PreSpawnStatus.set(PreSpawnStatus.FAIL);
             return false;
         } else {
-            return !SpawnPlacements.checkSpawnRules(entitytype, p_220422_, MobSpawnType.NATURAL, p_220427_, p_220422_.random)
-                ? false
-                : p_220422_.noCollision(entitytype.getSpawnAABB((double)p_220427_.getX() + 0.5, (double)p_220427_.getY(), (double)p_220427_.getZ() + 0.5));
+            if (!SpawnPlacements.checkSpawnRules(entitytype, p_220422_, MobSpawnType.NATURAL, p_220427_, p_220422_.random)) {
+                isValidSpawnPostitionForType$PreSpawnStatus.set(PreSpawnStatus.FAIL);
+                return false;
+            } else {
+                if (p_220422_.noCollision(entitytype.getSpawnAABB((double) p_220427_.getX() + 0.5, (double) p_220427_.getY(), (double) p_220427_.getZ() + 0.5))) {
+                    isValidSpawnPostitionForType$PreSpawnStatus.set(PreSpawnStatus.SUCCESS);
+                    return true;
+                } else {
+                    isValidSpawnPostitionForType$PreSpawnStatus.set(PreSpawnStatus.FAIL);
+                    return false;
+                }
+            }
+
         }
     }
 
@@ -272,7 +_,7 @@
         return p_46994_ > (double)(p_46993_.getType().getCategory().getDespawnDistance() * p_46993_.getType().getCategory().getDespawnDistance())
                 && p_46993_.removeWhenFarAway(p_46994_)
             ? false
-            : p_46993_.checkSpawnRules(p_46992_, MobSpawnType.NATURAL) && p_46993_.checkSpawnObstruction(p_46992_);
+            : net.neoforged.neoforge.event.EventHooks.checkSpawnPosition(p_46993_, p_46992_, MobSpawnType.NATURAL);
     }
 
     private static Optional<MobSpawnSettings.SpawnerData> getRandomSpawnMobAt(
@@ -303,9 +_,14 @@
         BlockPos p_220448_,
         @Nullable Holder<Biome> p_220449_
     ) {
-        return isInNetherFortressBounds(p_220448_, p_220444_, p_220447_, p_220445_)
-            ? NetherFortressStructure.FORTRESS_ENEMIES
-            : p_220446_.getMobsAt(p_220449_ != null ? p_220449_ : p_220444_.getBiome(p_220448_), p_220445_, p_220447_, p_220448_);
+        // Forge: Add in potential spawns, and replace hardcoded nether fortress mob list
+        if (isInNetherFortressBounds(p_220448_, p_220444_, p_220447_, p_220445_)) {
+            var monsterSpawns = p_220445_.registryAccess().registryOrThrow(Registries.STRUCTURE).getOrThrow(BuiltinStructures.FORTRESS).spawnOverrides().get(MobCategory.MONSTER);
+            if (monsterSpawns != null) { // structure modifiers can clear the spawn overrides
+                return net.neoforged.neoforge.event.EventHooks.getPotentialSpawns(p_220444_, p_220447_, p_220448_, monsterSpawns.spawns());
+            }
+        }
+        return net.neoforged.neoforge.event.EventHooks.getPotentialSpawns(p_220444_, p_220447_, p_220448_, p_220446_.getMobsAt(p_220449_ != null ? p_220449_ : p_220444_.getBiome(p_220448_), p_220445_, p_220447_, p_220448_));
     }
 
     public static boolean isInNetherFortressBounds(BlockPos p_220456_, ServerLevel p_220457_, MobCategory p_220458_, StructureManager p_220459_) {
@@ -392,11 +_,11 @@
 
                                 entity.moveTo(d0, (double)blockpos.getY(), d1, p_220454_.nextFloat() * 360.0F, 0.0F);
                                 if (entity instanceof Mob mob
-                                    && mob.checkSpawnRules(p_220451_, MobSpawnType.CHUNK_GENERATION)
-                                    && mob.checkSpawnObstruction(p_220451_)) {
+                                    && net.neoforged.neoforge.event.EventHooks.checkSpawnPosition(mob, p_220451_, MobSpawnType.CHUNK_GENERATION)) {
                                     spawngroupdata = mob.finalizeSpawn(
                                         p_220451_, p_220451_.getCurrentDifficultyAt(mob.blockPosition()), MobSpawnType.CHUNK_GENERATION, spawngroupdata
                                     );
+                                    mob.spawnReason(CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit
                                     p_220451_.addFreshEntityWithPassengers(mob);
                                     flag = true;
                                 }
@@ -516,9 +_,16 @@
             return this.unmodifiableMobCategoryCounts;
         }
 
+        int canSpawnForCategory$limit = -1;
         boolean canSpawnForCategory(MobCategory p_186549_, ChunkPos p_186550_) {
-            int i = p_186549_.getMaxInstancesPerChunk() * this.spawnableChunkCount / NaturalSpawner.MAGIC_NUMBER;
+            int i = (canSpawnForCategory$limit == -1 ? p_186549_.getMaxInstancesPerChunk() : canSpawnForCategory$limit) * this.spawnableChunkCount / NaturalSpawner.MAGIC_NUMBER;
+            canSpawnForCategory$limit = -1;
             return this.mobCategoryCounts.getInt(p_186549_) >= i ? false : this.localMobCapCalculator.canSpawn(p_186549_, p_186550_);
+        }
+
+        boolean canSpawnForCategory(MobCategory p_186549_, ChunkPos p_186550_, int limit) {
+            canSpawnForCategory$limit = limit;
+            return canSpawnForCategory(p_186549_, p_186550_);
         }
     }
 }
