--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -14,6 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
+import net.minecraft.core.QuartPos;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.server.level.ServerLevel;
@@ -25,8 +_,6 @@
 import net.minecraft.util.profiling.Profiler;
 import net.minecraft.util.profiling.ProfilerFiller;
 import net.minecraft.util.random.WeightedList;
-import net.minecraft.world.attribute.EnvironmentAttributeReader;
-import net.minecraft.world.attribute.EnvironmentAttributes;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySpawnReason;
 import net.minecraft.world.entity.EntityType;
@@ -34,6 +_,7 @@
 import net.minecraft.world.entity.MobCategory;
 import net.minecraft.world.entity.SpawnGroupData;
 import net.minecraft.world.entity.SpawnPlacements;
+import net.minecraft.world.entity.animal.feline.Ocelot;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.MobSpawnSettings;
@@ -50,6 +_,11 @@
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.storage.LevelData;
 import net.minecraft.world.phys.Vec3;
+import net.neoforged.neoforge.event.EventHooks;
+import org.bukkit.craftbukkit.util.CraftSpawnCategory;
+import org.bukkit.entity.Ageable;
+import org.bukkit.entity.SpawnCategory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -66,29 +_,51 @@
     }
 
     public static NaturalSpawner.SpawnState createState(
-        int spawnableChunkCount, ServerLevel level, NaturalSpawner.ChunkGetter chunkGetter, LocalMobCapCalculator localMobCapCalculator
-    ) {
-        Iterable<Entity> entities = level.getAllEntities();
+        int spawnableChunkCount, Iterable<Entity> entities, NaturalSpawner.ChunkGetter chunkGetter, LocalMobCapCalculator localMobCapCalculator
+    ) {
+        // Paper start - Optional per player mob spawns
+        return createState(spawnableChunkCount, entities, chunkGetter, localMobCapCalculator, false);
+    }
+
+    public static NaturalSpawner.SpawnState createState(
+            final int spawnableChunkCount,
+            final Iterable<Entity> entities,
+            final NaturalSpawner.ChunkGetter chunkGetter,
+            final LocalMobCapCalculator localMobCapCalculator,
+            final boolean countMobs
+    ) {
+        // Paper end - Optional per player mob spawns
         PotentialCalculator spawnPotential = new PotentialCalculator();
         Object2IntOpenHashMap<MobCategory> mobCounts = new Object2IntOpenHashMap<>();
 
         for (Entity entity : entities) {
             if (!(entity instanceof Mob mob && (mob.isPersistenceRequired() || mob.requiresCustomPersistence()))) {
-                MobCategory category = entity.getType().getCategory();
+                MobCategory category = entity.getClassification(true);
                 if (category != MobCategory.MISC) {
+                    // Paper start - Only count natural spawns
+                    if (!entity.level().paperConfig().entities.spawning.countAllMobsForSpawning &&
+                            !(entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL ||
+                                    entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN)) {
+                        continue;
+                    }
+                    // Paper end - Only count natural spawns
                     BlockPos pos = entity.blockPosition();
                     chunkGetter.query(ChunkPos.pack(pos), chunk -> {
-                        MobSpawnSettings naturalMobSpawns = level.environmentAttributes().getValue(EnvironmentAttributes.NATURAL_MOB_SPAWNS, pos);
-                        MobSpawnSettings.MobSpawnCost mobSpawnCost = naturalMobSpawns.getMobSpawnCost(entity.getType());
+                        MobSpawnSettings.MobSpawnCost mobSpawnCost = getRoughBiome(pos, chunk).getMobSettings().getMobSpawnCost(entity.getType());
                         if (mobSpawnCost != null) {
                             spawnPotential.addCharge(entity.blockPosition(), mobSpawnCost.charge());
                         }
 
-                        if (entity instanceof Mob) {
+                        if (localMobCapCalculator != null && entity instanceof Mob) { // Paper - Optional per player mob spawns
                             localMobCapCalculator.addMob(chunk.getPos(), category);
                         }
 
                         mobCounts.addTo(category, 1);
+                        // Paper start - Optional per player mob spawns
+                        if (countMobs) {
+                            ((ServerLevel) chunk.getLevel()).getChunkSource().chunkMap.updatePlayerMobTypeMap(entity);
+                        }
+                        // Paper end - Optional per player mob spawns
                     });
                 }
             }
@@ -97,43 +_,117 @@
         return new NaturalSpawner.SpawnState(spawnableChunkCount, mobCounts, spawnPotential, localMobCapCalculator);
     }
 
-    public static List<MobCategory> getFilteredSpawningCategories(NaturalSpawner.SpawnState state, boolean spawnEnemies, boolean spawnPersistent) {
+    private static Biome getRoughBiome(BlockPos pos, ChunkAccess chunk) {
+        return chunk.getNoiseBiome(QuartPos.fromBlock(pos.getX()), QuartPos.fromBlock(pos.getY()), QuartPos.fromBlock(pos.getZ())).value();
+    }
+
+    // CraftBukkit start - add server
+    public static List<MobCategory> getFilteredSpawningCategories(
+            final NaturalSpawner.SpawnState state, final boolean spawnFriendlies, final boolean spawnEnemies, final boolean spawnPersistent, final ServerLevel level
+    ) {
+        net.minecraft.world.level.storage.LevelData worlddata = level.getLevelData(); // CraftBukkit - Other mob type spawn tick rate
+        // CraftBukkit end
         List<MobCategory> spawningCategories = new ArrayList<>(SPAWNING_CATEGORIES.length);
 
         for (MobCategory mobCategory : SPAWNING_CATEGORIES) {
-            if ((spawnEnemies || mobCategory.isFriendly()) && (spawnPersistent || !mobCategory.isPersistent()) && state.canSpawnForCategoryGlobal(mobCategory)) {
+            // CraftBukkit start - Use per-world spawn limits
+            boolean spawnThisTick = true;
+            int limit = mobCategory.getMaxInstancesPerChunk();
+            org.bukkit.entity.SpawnCategory spawnCategory = org.bukkit.craftbukkit.util.CraftSpawnCategory.toBukkit(mobCategory);
+            if (org.bukkit.craftbukkit.util.CraftSpawnCategory.isValidForLimits(spawnCategory)) {
+                spawnThisTick = level.ticksPerSpawnCategory.getLong(spawnCategory) != 0 && worlddata.getGameTime() % level.ticksPerSpawnCategory.getLong(spawnCategory) == 0;
+                limit = level.getWorld().getSpawnLimit(spawnCategory);
+            }
+
+            if (!spawnThisTick || limit == 0) {
+                continue;
+            }
+
+            if ((spawnFriendlies || !mobCategory.isFriendly()) && (spawnEnemies || mobCategory.isFriendly()) && (spawnPersistent || !mobCategory.isPersistent()) && (level.paperConfig().entities.spawning.perPlayerMobSpawns || state.canSpawnForCategoryGlobal(mobCategory, limit))) { // Paper - Optional per player mob spawns; remove global check, check later during the local one
+                // CraftBukkit end
                 spawningCategories.add(mobCategory);
             }
         }
 
         return spawningCategories;
     }
-
     public static void spawnForChunk(ServerLevel level, LevelChunk chunk, NaturalSpawner.SpawnState state, List<MobCategory> spawningCategories) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("spawner");
 
         for (MobCategory mobCategory : spawningCategories) {
-            if (state.canSpawnForCategoryLocal(mobCategory, chunk.getPos())) {
-                spawnCategoryForChunk(mobCategory, level, chunk, state::canSpawn, state::afterSpawn);
+            // Paper start - Optional per player mob spawns
+            final boolean canSpawn;
+            int maxSpawns = Integer.MAX_VALUE;
+            if (level.paperConfig().entities.spawning.perPlayerMobSpawns) {
+                // Copied from getFilteredSpawningCategories
+                int limit = mobCategory.getMaxInstancesPerChunk();
+                org.bukkit.entity.SpawnCategory spawnCategory = org.bukkit.craftbukkit.util.CraftSpawnCategory.toBukkit(mobCategory);
+                if (org.bukkit.craftbukkit.util.CraftSpawnCategory.isValidForLimits(spawnCategory)) {
+                    limit = level.getWorld().getSpawnLimit(spawnCategory);
+                }
+
+                // Apply per-player limit
+                int minDiff = Integer.MAX_VALUE;
+                final ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.server.level.ServerPlayer> inRange =
+                        level.moonrise$getNearbyPlayers().getPlayers(chunk.getPos(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+                if (inRange != null) {
+                    final net.minecraft.server.level.ServerPlayer[] backingSet = inRange.getRawDataUnchecked();
+                    for (int k = 0, len = inRange.size(); k < len; k++) {
+                        minDiff = Math.min(limit - level.getChunkSource().chunkMap.getMobCountNear(backingSet[k], mobCategory), minDiff);
+                    }
+                }
+
+                maxSpawns = (minDiff == Integer.MAX_VALUE) ? 0 : minDiff;
+                canSpawn = maxSpawns > 0;
+            } else {
+                canSpawn = state.canSpawnForCategoryLocal(mobCategory, chunk.getPos());
+            }
+            if (canSpawn) {
+                spawnCategoryForChunk(mobCategory, level, chunk, state::canSpawn, state::afterSpawn,
+                        maxSpawns, level.paperConfig().entities.spawning.perPlayerMobSpawns ? level.getChunkSource().chunkMap::updatePlayerMobTypeMap : null);
+                // Paper end - Optional per player mob spawns
             }
         }
 
         profiler.pop();
     }
 
+    // Paper start - Add mobcaps commands
+    public static int globalLimitForCategory(final ServerLevel level, final MobCategory category, final int spawnableChunkCount) {
+        final int categoryLimit = level.getWorld().getSpawnLimitUnsafe(org.bukkit.craftbukkit.util.CraftSpawnCategory.toBukkit(category));
+        if (categoryLimit < 1) {
+            return categoryLimit;
+        }
+        return categoryLimit * spawnableChunkCount / NaturalSpawner.MAGIC_NUMBER;
+    }
+    // Paper end - Add mobcaps commands
+
     public static void spawnCategoryForChunk(
         MobCategory mobCategory, ServerLevel level, LevelChunk chunk, NaturalSpawner.SpawnPredicate extraTest, NaturalSpawner.AfterSpawnCallback spawnCallback
     ) {
+        // Paper start - Optional per player mob spawns
+        spawnCategoryForChunk(mobCategory, level, chunk, extraTest, spawnCallback, Integer.MAX_VALUE, null);
+    }
+    public static void spawnCategoryForChunk(
+            final MobCategory mobCategory,
+            final ServerLevel level,
+            final LevelChunk chunk,
+            final NaturalSpawner.SpawnPredicate extraTest,
+            final NaturalSpawner.AfterSpawnCallback spawnCallback,
+            final int maxSpawns,
+            final Consumer<Entity> trackEntity
+    ) {
+        // Paper end - Optional per player mob spawns
         BlockPos start = getRandomPosWithin(level, chunk);
         if (start.getY() >= level.getMinY() + 1) {
-            spawnCategoryForPosition(mobCategory, level, chunk, start, extraTest, spawnCallback);
+            spawnCategoryForPosition(mobCategory, level, chunk, start, extraTest, spawnCallback, maxSpawns, trackEntity); // Paper - Optional per player mob spawns
         }
     }
 
     @VisibleForDebug
     public static void spawnCategoryForPosition(MobCategory mobCategory, ServerLevel level, BlockPos start) {
-        spawnCategoryForPosition(mobCategory, level, level.getChunk(start), start, (var0, var1, var2, var3) -> true, (var0, var1) -> {});
+        spawnCategoryForPosition(mobCategory, level, level.getChunk(start), start, (type, chunk, pos) -> true, (mob, chunk) -> {});
     }
 
     public static void spawnCategoryForPosition(
@@ -144,11 +_,25 @@
         NaturalSpawner.SpawnPredicate extraTest,
         NaturalSpawner.AfterSpawnCallback spawnCallback
     ) {
+        // Paper start - Optional per player mob spawns
+        spawnCategoryForPosition(mobCategory, level, chunk, start, extraTest, spawnCallback, Integer.MAX_VALUE, null);
+    }
+    public static void spawnCategoryForPosition(
+            final MobCategory mobCategory,
+            final ServerLevel level,
+            final ChunkAccess chunk,
+            final BlockPos start,
+            final NaturalSpawner.SpawnPredicate extraTest,
+            final NaturalSpawner.AfterSpawnCallback spawnCallback,
+            final int maxSpawns,
+            final @Nullable Consumer<Entity> trackEntity
+    ) {
+        // Paper end - Optional per player mob spawns
         StructureManager structureManager = level.structureManager();
         ChunkGenerator generator = level.getChunkSource().getGenerator();
         int yStart = start.getY();
-        BlockState state = chunk.getBlockState(start);
-        if (!state.isRedstoneConductor(chunk, start)) {
+        BlockState state = level.getBlockStateIfLoadedAndInBounds(start); // Paper - don't load chunks for mob spawn
+        if (state != null && !state.isRedstoneConductor(chunk, start)) { // Paper - don't load chunks for mob spawn
             BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
             int clusterSize = 0;
 
@@ -167,10 +_,10 @@
                     pos.set(x, yStart, z);
                     double xx = x + 0.5;
                     double zz = z + 0.5;
-                    Player nearestPlayer = level.getNearestPlayer(xx, yStart, zz, -1.0, false);
+                    Player nearestPlayer = level.getNearestPlayer(xx, yStart, zz, -1.0, level.purpurConfig.mobSpawningIgnoreCreativePlayers); // Purpur - mob spawning option to ignore creative players
                     if (nearestPlayer != null) {
                         double nearestPlayerDistanceSqr = nearestPlayer.distanceToSqr(xx, yStart, zz);
-                        if (isRightDistanceToPlayerAndSpawnPoint(level, chunk, pos, nearestPlayerDistanceSqr)) {
+                        if (level.isLoadedAndInBounds(pos) && isRightDistanceToPlayerAndSpawnPoint(level, chunk, pos, nearestPlayerDistanceSqr)) { // Paper - don't load chunks for mob spawn
                             if (currentSpawnData == null) {
                                 Optional<MobSpawnSettings.SpawnerData> nextSpawnData = getRandomSpawnMobAt(
                                     level, structureManager, generator, mobCategory, level.random, pos
@@ -180,11 +_,22 @@
                                 }
 
                                 currentSpawnData = nextSpawnData.get();
-                                max = currentSpawnData.count().sample(level.random);
+                                max = currentSpawnData.minCount() + level.random.nextInt(1 + currentSpawnData.maxCount() - currentSpawnData.minCount());
                             }
 
-                            if (isValidSpawnPostitionForType(level, mobCategory, structureManager, generator, currentSpawnData, pos, nearestPlayerDistanceSqr)
-                                && extraTest.test(currentSpawnData.type(), level, pos, chunk)) {
+                            // Paper start - PreCreatureSpawnEvent
+                            PreSpawnStatus doSpawning = isValidSpawnPostitionForType(level, mobCategory, structureManager, generator, currentSpawnData, pos, nearestPlayerDistanceSqr);
+                            // Paper start - per player mob count backoff
+                            if (doSpawning == PreSpawnStatus.ABORT || doSpawning == PreSpawnStatus.CANCELLED) {
+                                level.getChunkSource().chunkMap.updateFailurePlayerMobTypeMap(pos.getX() >> 4, pos.getZ() >> 4, mobCategory);
+                            }
+                            // Paper end - per player mob count backoff
+                            if (doSpawning == PreSpawnStatus.ABORT) {
+                                return;
+                            }
+                            if (doSpawning == PreSpawnStatus.SUCCESS
+                                    // Paper end - PreCreatureSpawnEvent
+                                && extraTest.test(currentSpawnData.type(), pos, chunk)) {
                                 Mob mob = getMobForSpawn(level, currentSpawnData.type());
                                 if (mob == null) {
                                     return;
@@ -195,11 +_,19 @@
                                     groupData = mob.finalizeSpawn(
                                         level, level.getCurrentDifficultyAt(mob.blockPosition()), EntitySpawnReason.NATURAL, groupData
                                     );
-                                    clusterSize++;
-                                    groupSize++;
+                                    mob.spawnReason((mob instanceof Ocelot && !((Ageable) mob.getBukkitEntity()).isAdult()) ? CreatureSpawnEvent.SpawnReason.OCELOT_BABY : CreatureSpawnEvent.SpawnReason.NATURAL);
                                     level.addFreshEntityWithPassengers(mob);
-                                    spawnCallback.run(mob, chunk);
-                                    if (clusterSize >= mob.getMaxSpawnClusterSize()) {
+                                    if (!mob.isRemoved()) {
+                                        clusterSize++;
+                                        groupSize++;
+                                        spawnCallback.run(mob, chunk);
+                                        // Paper start - Optional per player mob spawns
+                                        if (trackEntity != null) {
+                                            trackEntity.accept(mob);
+                                        }
+                                        // Paper end - Optional per player mob spawns
+                                    }
+                                    if (clusterSize >= EventHooks.getMaxSpawnClusterSize(mob)) {
                                         return;
                                     }
 
@@ -232,7 +_,15 @@
         return Objects.equals(chunkPos, chunk.getPos()) || level.canSpawnEntitiesInChunk(chunkPos);
     }
 
-    private static boolean isValidSpawnPostitionForType(
+    // Paper start - PreCreatureSpawnEvent
+    private enum PreSpawnStatus {
+        FAIL,
+        SUCCESS,
+        CANCELLED,
+        ABORT
+    }
+    private static PreSpawnStatus isValidSpawnPostitionForType(
+        // Paper end - PreCreatureSpawnEvent
         ServerLevel level,
         MobCategory mobCategory,
         StructureManager structureManager,
@@ -241,21 +_,27 @@
         BlockPos.MutableBlockPos pos,
         double nearestPlayerDistanceSqr
     ) {
+        // Paper start - PreCreatureSpawnEvent
         EntityType<?> type = currentSpawnData.type();
-        if (type.getCategory() == MobCategory.MISC) {
-            return false;
-        } else if (!type.canSpawnFarFromPlayer()
-            && nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()) {
-            return false;
-        } else if (!type.canSummon() || !canSpawnMobAt(level, structureManager, generator, mobCategory, currentSpawnData, pos)) {
-            return false;
-        } else if (!SpawnPlacements.isSpawnPositionOk(type, level, pos)) {
-            return false;
-        } else {
-            return !SpawnPlacements.checkSpawnRules(type, level, EntitySpawnReason.NATURAL, pos, level.random)
-                ? false
-                : level.noCollision(type.getSpawnAABB(pos.getX() + 0.5, pos.getY(), pos.getZ() + 0.5));
+        com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
+                org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level),
+                org.bukkit.craftbukkit.entity.CraftEntityType.minecraftToBukkit(type), org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL
+        );
+        if (!event.callEvent()) {
+            if (event.shouldAbortSpawn()) {
+                return PreSpawnStatus.ABORT;
+            }
+            return PreSpawnStatus.CANCELLED;
         }
+        final boolean success = type.getCategory() != MobCategory.MISC
+                // Paper end - PreCreatureSpawnEvent
+                && (type.canSpawnFarFromPlayer() || !(nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()))
+                && type.canSummon()
+                && canSpawnMobAt(level, structureManager, generator, mobCategory, currentSpawnData, pos)
+                && SpawnPlacements.isSpawnPositionOk(type, level, pos)
+                && SpawnPlacements.checkSpawnRules(type, level, EntitySpawnReason.NATURAL, pos, level.random)
+                && level.noCollision(type.getSpawnAABB(pos.getX() + 0.5, pos.getY(), pos.getZ() + 0.5));
+        return success ? PreSpawnStatus.SUCCESS : PreSpawnStatus.FAIL; // Paper - PreCreatureSpawnEvent
     }
 
     private static @Nullable Mob getMobForSpawn(ServerLevel level, EntityType<?> type) {
@@ -267,6 +_,7 @@
             LOGGER.warn("Can't spawn entity of type: {}", BuiltInRegistries.ENTITY_TYPE.getKey(type));
         } catch (Exception e) {
             LOGGER.warn("Failed to create mob", e);
+            com.destroystokyo.paper.exception.ServerInternalException.reportInternalException(e); // Paper - ServerExceptionEvent
         }
 
         return null;
@@ -276,7 +_,7 @@
         return nearestPlayerDistanceSqr > mob.getType().getCategory().getDespawnDistance() * mob.getType().getCategory().getDespawnDistance()
                 && mob.removeWhenFarAway(nearestPlayerDistanceSqr)
             ? false
-            : mob.checkSpawnRules(level, EntitySpawnReason.NATURAL) && mob.checkSpawnObstruction(level);
+            : net.neoforged.neoforge.event.EventHooks.checkSpawnPosition(mob, level, EntitySpawnReason.NATURAL);
     }
 
     private static Optional<MobSpawnSettings.SpawnerData> getRandomSpawnMobAt(
@@ -285,7 +_,7 @@
         Holder<Biome> biome = level.getBiome(pos);
         return mobCategory == MobCategory.WATER_AMBIENT && biome.is(BiomeTags.REDUCED_WATER_AMBIENT_SPAWNS) && random.nextFloat() < 0.98F
             ? Optional.empty()
-            : mobsAt(level, structureManager, generator, mobCategory, pos).getRandom(random);
+            : mobsAt(level, structureManager, generator, mobCategory, pos, biome).getRandom(random);
     }
 
     private static boolean canSpawnMobAt(
@@ -296,15 +_,20 @@
         MobSpawnSettings.SpawnerData spawnerData,
         BlockPos pos
     ) {
-        return mobsAt(level, structureManager, generator, mobCategory, pos).contains(spawnerData);
+        return mobsAt(level, structureManager, generator, mobCategory, pos, null).contains(spawnerData);
     }
 
     private static WeightedList<MobSpawnSettings.SpawnerData> mobsAt(
-        ServerLevel level, StructureManager structureManager, ChunkGenerator generator, MobCategory mobCategory, BlockPos pos
+        ServerLevel level, StructureManager structureManager, ChunkGenerator generator, MobCategory mobCategory, BlockPos pos, @Nullable Holder<Biome> biome
     ) {
-        return isInNetherFortressBounds(pos, level, mobCategory, structureManager)
-            ? NetherFortressStructure.FORTRESS_ENEMIES
-            : generator.getMobsAt(level, structureManager, mobCategory, pos);
+        // Forge: Add in potential spawns, and replace hardcoded nether fortress mob list
+        if (isInNetherFortressBounds(pos, level, mobCategory, structureManager)) {
+            var monsterSpawns = structureManager.registryAccess().lookupOrThrow(Registries.STRUCTURE).getValueOrThrow(BuiltinStructures.FORTRESS).spawnOverrides().get(MobCategory.MONSTER);
+            if (monsterSpawns != null) { // structure modifiers can clear the spawn overrides
+                return net.neoforged.neoforge.event.EventHooks.getPotentialSpawns(level, mobCategory, pos, monsterSpawns.spawns());
+            }
+        }
+        return net.neoforged.neoforge.event.EventHooks.getPotentialSpawns(level, mobCategory, pos, generator.getMobsAt(biome != null ? biome : level.getBiome(pos), structureManager, mobCategory, pos));
     }
 
     public static boolean isInNetherFortressBounds(BlockPos pos, ServerLevel level, MobCategory category, StructureManager structureManager) {
@@ -337,79 +_,75 @@
         }
     }
 
-    public static void spawnMobsForChunkGeneration(ServerLevelAccessor level, BlockPos sourcePos, ChunkPos chunkPos, RandomSource random) {
-        if (level.getLevel().getGameRules().get(GameRules.SPAWN_MOBS)) {
-            EnvironmentAttributeReader attributes = level.environmentAttributes();
-            MobSpawnSettings mobSettings = attributes.getValue(EnvironmentAttributes.NATURAL_MOB_SPAWNS, sourcePos);
-            WeightedList<MobSpawnSettings.SpawnerData> mobs = mobSettings.getMobsToSpawn(MobCategory.CREATURE);
-            if (!mobs.isEmpty()) {
-                float creatureProbability = attributes.getValue(EnvironmentAttributes.CREATURE_WORLD_GEN_SPAWN_PROBABILITY, sourcePos);
-                int xo = chunkPos.getMinBlockX();
-                int zo = chunkPos.getMinBlockZ();
-
-                while (random.nextFloat() < creatureProbability) {
-                    Optional<MobSpawnSettings.SpawnerData> nextSpawnerData = mobs.getRandom(random);
-                    if (!nextSpawnerData.isEmpty()) {
-                        MobSpawnSettings.SpawnerData spawnerData = nextSpawnerData.get();
-                        int count = spawnerData.count().sample(random);
-                        SpawnGroupData groupSpawnData = null;
-                        int x = xo + random.nextInt(16);
-                        int z = zo + random.nextInt(16);
-                        int startX = x;
-                        int startZ = z;
-
-                        for (int i = 0; i < count; i++) {
-                            boolean success = false;
-
-                            for (int attempts = 0; !success && attempts < 4; attempts++) {
-                                BlockPos pos = getTopNonCollidingPos(level, spawnerData.type(), x, z);
-                                if (spawnerData.type().canSummon() && SpawnPlacements.isSpawnPositionOk(spawnerData.type(), level, pos)) {
-                                    float width = spawnerData.type().getWidth();
-                                    double fx = Mth.clamp(x, (double)xo + width, xo + 16.0 - width);
-                                    double fz = Mth.clamp(z, (double)zo + width, zo + 16.0 - width);
-                                    if (!level.noCollision(spawnerData.type().getSpawnAABB(fx, pos.getY(), fz))
-                                        || !SpawnPlacements.checkSpawnRules(
-                                            spawnerData.type(),
-                                            level,
-                                            EntitySpawnReason.CHUNK_GENERATION,
-                                            BlockPos.containing(fx, pos.getY(), fz),
-                                            level.getRandom()
-                                        )) {
-                                        continue;
-                                    }
-
-                                    Entity entity;
-                                    try {
-                                        entity = spawnerData.type().create(level.getLevel(), EntitySpawnReason.NATURAL);
-                                    } catch (Exception e) {
-                                        LOGGER.warn("Failed to create mob", e);
-                                        continue;
-                                    }
-
-                                    if (entity == null) {
-                                        continue;
-                                    }
-
-                                    entity.snapTo(fx, pos.getY(), fz, random.nextFloat() * 360.0F, 0.0F);
-                                    if (entity instanceof Mob mob
-                                        && mob.checkSpawnRules(level, EntitySpawnReason.CHUNK_GENERATION)
-                                        && mob.checkSpawnObstruction(level)) {
-                                        groupSpawnData = mob.finalizeSpawn(
-                                            level, level.getCurrentDifficultyAt(mob.blockPosition()), EntitySpawnReason.CHUNK_GENERATION, groupSpawnData
-                                        );
-                                        level.addFreshEntityWithPassengers(mob);
-                                        success = true;
-                                    }
-                                }
-
-                                x += random.nextInt(5) - random.nextInt(5);
-
-                                for (z += random.nextInt(5) - random.nextInt(5);
-                                    x < xo || x >= xo + 16 || z < zo || z >= zo + 16;
-                                    z = startZ + random.nextInt(5) - random.nextInt(5)
-                                ) {
-                                    x = startX + random.nextInt(5) - random.nextInt(5);
-                                }
+    public static void spawnMobsForChunkGeneration(ServerLevelAccessor level, Holder<Biome> biome, ChunkPos chunkPos, RandomSource random) {
+        MobSpawnSettings mobSettings = biome.value().getMobSettings();
+        WeightedList<MobSpawnSettings.SpawnerData> mobs = mobSettings.getMobs(MobCategory.CREATURE);
+        if (!mobs.isEmpty() && level.getLevel().getGameRules().get(GameRules.SPAWN_MOBS)) {
+            int xo = chunkPos.getMinBlockX();
+            int zo = chunkPos.getMinBlockZ();
+
+            while (random.nextFloat() < mobSettings.getCreatureProbability()) {
+                Optional<MobSpawnSettings.SpawnerData> nextSpawnerData = mobs.getRandom(random);
+                if (!nextSpawnerData.isEmpty()) {
+                    MobSpawnSettings.SpawnerData spawnerData = nextSpawnerData.get();
+                    int count = spawnerData.minCount() + random.nextInt(1 + spawnerData.maxCount() - spawnerData.minCount());
+                    SpawnGroupData groupSpawnData = null;
+                    int x = xo + random.nextInt(16);
+                    int z = zo + random.nextInt(16);
+                    int startX = x;
+                    int startZ = z;
+
+                    for (int i = 0; i < count; i++) {
+                        boolean success = false;
+
+                        for (int attempts = 0; !success && attempts < 4; attempts++) {
+                            BlockPos pos = getTopNonCollidingPos(level, spawnerData.type(), x, z);
+                            if (spawnerData.type().canSummon() && SpawnPlacements.isSpawnPositionOk(spawnerData.type(), level, pos)) {
+                                float width = spawnerData.type().getWidth();
+                                double fx = Mth.clamp(x, (double)xo + width, xo + 16.0 - width);
+                                double fz = Mth.clamp(z, (double)zo + width, zo + 16.0 - width);
+                                if (!level.noCollision(spawnerData.type().getSpawnAABB(fx, pos.getY(), fz))
+                                    || !SpawnPlacements.checkSpawnRules(
+                                        spawnerData.type(),
+                                        level,
+                                        EntitySpawnReason.CHUNK_GENERATION,
+                                        BlockPos.containing(fx, pos.getY(), fz),
+                                        level.getRandom()
+                                    )) {
+                                    continue;
+                                }
+
+                                Entity entity;
+                                try {
+                                    entity = spawnerData.type().create(level.getLevel(), EntitySpawnReason.NATURAL);
+                                } catch (Exception e) {
+                                    LOGGER.warn("Failed to create mob", e);
+                                    com.destroystokyo.paper.exception.ServerInternalException.reportInternalException(e); // Paper - ServerExceptionEvent
+                                    continue;
+                                }
+
+                                if (entity == null) {
+                                    continue;
+                                }
+
+                                entity.snapTo(fx, pos.getY(), fz, random.nextFloat() * 360.0F, 0.0F);
+                                if (entity instanceof Mob mob
+                                    && net.neoforged.neoforge.event.EventHooks.checkSpawnPosition(mob, level, EntitySpawnReason.CHUNK_GENERATION)) {
+                                    groupSpawnData = mob.finalizeSpawn(
+                                        level, level.getCurrentDifficultyAt(mob.blockPosition()), EntitySpawnReason.CHUNK_GENERATION, groupSpawnData
+                                    );
+                                    level.addFreshEntityWithPassengers(mob, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit
+                                    success = true;
+                                }
+                            }
+
+                            x += random.nextInt(5) - random.nextInt(5);
+
+                            for (z += random.nextInt(5) - random.nextInt(5);
+                                x < xo || x >= xo + 16 || z < zo || z >= zo + 16;
+                                z = startZ + random.nextInt(5) - random.nextInt(5)
+                            ) {
+                                x = startX + random.nextInt(5) - random.nextInt(5);
                             }
                         }
                     }
@@ -446,7 +_,7 @@
 
     @FunctionalInterface
     public interface SpawnPredicate {
-        boolean test(final EntityType<?> type, final ServerLevel level, final BlockPos blockPos, final ChunkAccess levelChunk);
+        boolean test(final EntityType<?> type, final BlockPos blockPos, final ChunkAccess levelChunk);
     }
 
     public static class SpawnState {
@@ -472,11 +_,10 @@
             this.unmodifiableMobCategoryCounts = Object2IntMaps.unmodifiable(mobCategoryCounts);
         }
 
-        private boolean canSpawn(EntityType<?> type, Level level, BlockPos testPos, ChunkAccess chunk) {
+        private boolean canSpawn(EntityType<?> type, BlockPos testPos, ChunkAccess chunk) {
             this.lastCheckedPos = testPos;
             this.lastCheckedType = type;
-            MobSpawnSettings naturalMobSpawns = level.environmentAttributes().getValue(EnvironmentAttributes.NATURAL_MOB_SPAWNS, testPos);
-            MobSpawnSettings.MobSpawnCost mobSpawnCost = naturalMobSpawns.getMobSpawnCost(type);
+            MobSpawnSettings.MobSpawnCost mobSpawnCost = NaturalSpawner.getRoughBiome(testPos, chunk).getMobSettings().getMobSpawnCost(type);
             if (mobSpawnCost == null) {
                 this.lastCharge = 0.0;
                 return true;
@@ -495,8 +_,7 @@
             if (pos.equals(this.lastCheckedPos) && type == this.lastCheckedType) {
                 charge = this.lastCharge;
             } else {
-                MobSpawnSettings naturalMobSpawns = mob.level().environmentAttributes().getValue(EnvironmentAttributes.NATURAL_MOB_SPAWNS, pos);
-                MobSpawnSettings.MobSpawnCost mobSpawnCost = naturalMobSpawns.getMobSpawnCost(type);
+                MobSpawnSettings.MobSpawnCost mobSpawnCost = NaturalSpawner.getRoughBiome(pos, chunk).getMobSettings().getMobSpawnCost(type);
                 if (mobSpawnCost != null) {
                     charge = mobSpawnCost.charge();
                 } else {
@@ -507,7 +_,7 @@
             this.spawnPotential.addCharge(pos, charge);
             MobCategory category = type.getCategory();
             this.mobCategoryCounts.addTo(category, 1);
-            this.localMobCapCalculator.addMob(ChunkPos.containing(pos), category);
+            if (this.localMobCapCalculator != null) this.localMobCapCalculator.addMob(ChunkPos.containing(pos), category); // Paper - Optional per player mob spawns
         }
 
         public int getSpawnableChunkCount() {
@@ -521,6 +_,13 @@
         private boolean canSpawnForCategoryGlobal(MobCategory mobCategory) {
             int maxMobCount = mobCategory.getMaxInstancesPerChunk() * this.spawnableChunkCount / NaturalSpawner.MAGIC_NUMBER;
             return this.mobCategoryCounts.getInt(mobCategory) < maxMobCount;
+        }
+
+        // CraftBukkit start
+        private boolean canSpawnForCategoryGlobal(MobCategory mobCategory, int limit) {
+            int i = limit * this.spawnableChunkCount / NaturalSpawner.MAGIC_NUMBER;
+            // CraftBukkit end
+            return this.mobCategoryCounts.getInt(mobCategory) < i;
         }
 
         private boolean canSpawnForCategoryLocal(MobCategory mobCategory, ChunkPos chunkPos) {
