--- a/net/minecraft/world/level/ServerLevelAccessor.java
+++ b/net/minecraft/world/level/ServerLevelAccessor.java
@@ -9,4 +_,13 @@
     default void addFreshEntityWithPassengers(Entity p_47206_) {
         p_47206_.getSelfAndPassengers().forEach(this::addFreshEntity);
     }
+
+    default void addFreshEntityWithPassengers(Entity p_47206_, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        p_47206_.getSelfAndPassengers().forEach((e) -> this.addFreshEntity(e, reason));
+    }
+
+    @Override
+    default ServerLevel getMinecraftWorld() {
+        return getLevel();
+    }
 }
