--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -36,6 +_,13 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import net.minecraft.world.entity.boss.enderdragon.EnderDragonPart;
+import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
+import net.minecraft.world.level.block.Blocks;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
+
 public class ServerExplosion implements Explosion {
     private static final ExplosionDamageCalculator EXPLOSION_DAMAGE_CALCULATOR = new ExplosionDamageCalculator();
     private static final int MAX_DROPS_PER_COMBINED_STACK = 16;
@@ -49,6 +_,10 @@
     private final DamageSource damageSource;
     private final ExplosionDamageCalculator damageCalculator;
     private final Map<Player, Vec3> hitPlayers = new HashMap<>();
+    // CraftBukkit - add field
+    public boolean wasCanceled = false;
+    public float yield;
+    // CraftBukkit end
 
     public ServerExplosion(
         ServerLevel level,
@@ -62,12 +_,13 @@
     ) {
         this.level = level;
         this.source = source;
-        this.radius = radius;
+        this.radius = (float) Math.max(radius, 0.0); // CraftBukkit - clamp bad values
         this.center = center;
         this.fire = fire;
         this.blockInteraction = blockInteraction;
         this.damageSource = damageSource == null ? level.damageSources().explosion(this) : damageSource;
         this.damageCalculator = damageCalculator == null ? this.makeDamageCalculator(source) : damageCalculator;
+        this.yield = this.blockInteraction == Explosion.BlockInteraction.DESTROY_WITH_DECAY ? 1.0F / this.radius : 1.0F; // CraftBukkit
     }
 
     private ExplosionDamageCalculator makeDamageCalculator(@Nullable Entity source) {
@@ -169,7 +_,12 @@
         return new ObjectArrayList<>(toBlowSet);
     }
 
+    @Deprecated
     private void hurtEntities() {
+        this.hurtEntities(List.of());
+    }
+
+    private void hurtEntities(List<BlockPos> blocks) {
         if (!(this.radius < 1.0E-5F)) {
             float doubleRadius = this.radius * 2.0F;
             int x0 = Mth.floor(this.center.x - doubleRadius - 1.0);
@@ -179,7 +_,9 @@
             int z0 = Mth.floor(this.center.z - doubleRadius - 1.0);
             int z1 = Mth.floor(this.center.z + doubleRadius + 1.0);
 
-            for (Entity entity : this.level.getEntities(this.source, new AABB(x0, y0, z0, x1, y1, z1))) {
+            List<Entity> list = this.level.getEntities(this.source, new AABB(x0, y0, z0, x1, y1, z1));
+            net.neoforged.neoforge.event.EventHooks.onExplosionDetonate(this.level, this, list, blocks);
+            for (Entity entity : list) {
                 if (!entity.ignoreExplosion(this)) {
                     double dist = Math.sqrt(entity.distanceToSqr(this.center)) / doubleRadius;
                     if (!(dist > 1.0)) {
@@ -188,15 +_,44 @@
                         boolean shouldDamageEntity = this.damageCalculator.shouldDamageEntity(this, entity);
                         float knockbackMultiplier = this.damageCalculator.getKnockbackMultiplier(entity);
                         float exposure = !shouldDamageEntity && knockbackMultiplier == 0.0F ? 0.0F : getSeenPercent(this.center, entity);
+                        // CraftBukkit start
                         if (shouldDamageEntity) {
-                            entity.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity, exposure));
+                            if (entity instanceof EnderDragonPart) {
+                                continue;
+                            }
+
+                            entity.lastDamageCancelled = false;
+
+                            if (entity instanceof EnderDragon) {
+                                for (EnderDragonPart entityComplexPart : ((EnderDragon) entity).subEntities) {
+                                    if (list.contains(entityComplexPart)) {
+                                        entityComplexPart.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity, exposure));
+                                    }
+                                }
+                            } else {
+                                entity.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity, exposure));
+                            }
+
+                            if (entity.lastDamageCancelled) {
+                                continue;
+                            }
                         }
+                        // CraftBukkit end
 
                         double knockbackResistance = entity instanceof LivingEntity livingEntity
                             ? livingEntity.getAttributeValue(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE)
                             : 0.0;
                         double knockbackPower = (1.0 - dist) * exposure * knockbackMultiplier * (1.0 - knockbackResistance);
                         Vec3 knockback = direction.scale(knockbackPower);
+                        knockback = net.neoforged.neoforge.event.EventHooks.getExplosionKnockback(this.level, this, entity, knockback, blocks);
+                        // CraftBukkit start - Call EntityKnockbackEvent
+                        if (entity instanceof LivingEntity) {
+                            Vec3 result = entity.getDeltaMovement().add(knockback);
+                            org.bukkit.event.entity.EntityKnockbackEvent event = CraftEventFactory.callEntityKnockbackEvent((org.bukkit.craftbukkit.entity.CraftLivingEntity) entity.getBukkitEntity(), source, org.bukkit.event.entity.EntityKnockbackEvent.KnockbackCause.EXPLOSION, knockbackPower, knockback, result.x, result.y, result.z);
+
+                            knockback = (event.isCancelled()) ? Vec3.ZERO : new Vec3(event.getFinalKnockback().getX(), event.getFinalKnockback().getY(), event.getFinalKnockback().getZ()).subtract(entity.getDeltaMovement());
+                        }
+                        // CraftBukkit end
                         entity.push(knockback);
                         if (entity.is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entity instanceof Projectile projectile) {
                             projectile.setOwner(this.damageSource.getEntity());
@@ -214,8 +_,29 @@
     private void interactWithBlocks(List<BlockPos> targetBlocks) {
         List<ServerExplosion.StackCollector> stacks = new ArrayList<>();
         Util.shuffle(targetBlocks, this.level.random);
+        // CraftBukkit start
+        List<org.bukkit.block.Block> bukkitBlocks = CraftEventFactory.handleExplodeEvent(this, targetBlocks);
+
+        targetBlocks.clear();
+        targetBlocks.addAll(bukkitBlocks.stream().map(bblock -> new BlockPos(bblock.getX(), bblock.getY(), bblock.getZ())).toList());
+
+        if (this.wasCanceled) {
+            return;
+        }
+        // CraftBukkit end
 
         for (BlockPos pos : targetBlocks) {
+            // CraftBukkit start - TNTPrimeEvent
+            BlockState blockstate = this.level.getBlockState(pos);
+            Block block = blockstate.getBlock();
+            if (block instanceof net.minecraft.world.level.block.TntBlock) {
+                BlockPos sourceBlock = source == null ? BlockPos.containing(this.center) : null;
+                if (!CraftEventFactory.callTNTPrimeEvent(this.level, pos, org.bukkit.event.block.TNTPrimeEvent.PrimeCause.EXPLOSION, source, sourceBlock)) {
+                    this.level.sendBlockUpdated(pos, Blocks.AIR.defaultBlockState(), blockstate, 3); // Update the block on the client
+                    continue;
+                }
+            }
+            // CraftBukkit end
             this.level.getBlockState(pos).onExplosionHit(this.level, pos, this, (stackx, position) -> addOrAppendStack(stacks, stackx, position));
         }
 
@@ -227,21 +_,38 @@
     private void createFire(List<BlockPos> targetBlocks) {
         for (BlockPos pos : targetBlocks) {
             if (this.level.random.nextInt(3) == 0 && this.level.getBlockState(pos).isAir() && this.level.getBlockState(pos.below()).isSolidRender()) {
-                this.level.setBlockAndUpdate(pos, BaseFireBlock.getState(this.level, pos));
+                // CraftBukkit start - Ignition by explosion
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(this.level, pos, this).isCancelled()) {
+                    this.level.setBlockAndUpdate(pos, BaseFireBlock.getState(this.level, pos));
+                }
+                // CraftBukkit end
             }
         }
     }
 
     public int explode() {
+        // CraftBukkit start
+        if (this.radius < 0.1F) {
+            return 0;
+        }
+        // CraftBukkit end
         this.level.gameEvent(this.source, GameEvent.EXPLODE, this.center);
         List<BlockPos> toBlow = this.calculateExplodedPositions();
-        this.hurtEntities();
+        this.hurtEntities(toBlow);
         if (this.interactsWithBlocks()) {
             ProfilerFiller profiler = Profiler.get();
             profiler.push("explosion_blocks");
             this.interactWithBlocks(toBlow);
             profiler.pop();
+            // CraftBukkit start - handle KEEP effect
+        } else {
+            Util.shuffle(toBlow, this.level.random); // CraftBukkit - Copy from calculateExplodedPositions
+            List<org.bukkit.block.Block> bukkitBlocks = CraftEventFactory.handleExplodeEvent(this, toBlow);
+
+            toBlow.clear();
+            toBlow.addAll(bukkitBlocks.stream().map(bblock -> new BlockPos(bblock.getX(), bblock.getY(), bblock.getZ())).toList());
         }
+        // CraftBukkit end
 
         if (this.fire) {
             this.createFire(toBlow);
@@ -251,6 +_,7 @@
     }
 
     private static void addOrAppendStack(List<ServerExplosion.StackCollector> stacks, ItemStack stack, BlockPos pos) {
+        if (stack.isEmpty()) return; // CraftBukkit - SPIGOT-5425
         for (ServerExplosion.StackCollector stackCollector : stacks) {
             stackCollector.tryMerge(stack);
             if (stack.isEmpty()) {
