--- a/net/minecraft/world/level/SignalGetter.java
+++ b/net/minecraft/world/level/SignalGetter.java
@@ -4,7 +_,7 @@
 import net.minecraft.core.Direction;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.DiodeBlock;
-import net.minecraft.world.level.block.RedstoneWireBlock;
+import net.minecraft.world.level.block.RedStoneWireBlock;
 import net.minecraft.world.level.block.state.BlockState;
 
 public interface SignalGetter extends BlockGetter {
@@ -52,7 +_,7 @@
         } else if (blockState.is(Blocks.REDSTONE_BLOCK)) {
             return 15;
         } else if (blockState.is(Blocks.REDSTONE_WIRE)) {
-            return blockState.getValue(RedstoneWireBlock.POWER);
+            return blockState.getValue(RedStoneWireBlock.POWER);
         } else {
             return blockState.isSignalSource() ? this.getDirectSignal(pos, direction) : 0;
         }
@@ -65,7 +_,7 @@
     default int getSignal(BlockPos pos, Direction direction) {
         BlockState state = this.getBlockState(pos);
         int signal = state.getSignal(this, pos, direction);
-        return state.isRedstoneConductor(this, pos) ? Math.max(signal, this.getDirectSignalTo(pos)) : signal;
+        return state.shouldCheckWeakPower(this, pos, direction) ? Math.max(signal, this.getDirectSignalTo(pos)) : signal;
     }
 
     default int getBestOwnOrNeighbourSignal(BlockPos pos) {
