--- a/net/minecraft/client/gui/render/GuiItemAtlas.java
+++ b/net/minecraft/client/gui/render/GuiItemAtlas.java
@@ -102,15 +_,7 @@
         if (clear) {
             device.createCommandEncoder()
                 .clearColorAndDepthTextures(
-                    this.texture,
-                    GuiRenderer.CLEAR_COLOR,
-                    this.depthTexture,
-                    0.0,
-                    left,
-                    this.textureSize - bottom,
-                    this.slotTextureSize,
-                    this.slotTextureSize,
-                    0
+                    this.texture, GuiRenderer.CLEAR_COLOR, this.depthTexture, 0.0, left, this.textureSize - bottom, this.slotTextureSize, this.slotTextureSize
                 );
         }
 
