--- a/net/minecraft/data/worldgen/placement/AquaticPlacements.java
+++ b/net/minecraft/data/worldgen/placement/AquaticPlacements.java
@@ -76,14 +_,6 @@
             BlockPredicateFilter.forPredicate(BlockPredicate.matchesBlocks(Blocks.WATER)),
             BiomeFilter.biome()
         );
-        BlockPredicateFilter kelpPlacementFilter = BlockPredicateFilter.forPredicate(
-            BlockPredicate.allOf(
-                BlockPredicate.matchesBlocks(Blocks.WATER),
-                BlockPredicate.matchesBlocks(Direction.UP, Blocks.WATER),
-                BlockPredicate.hasSturdyFace(Direction.DOWN, Direction.UP),
-                BlockPredicate.not(BlockPredicate.matchesTag(Direction.DOWN, BlockTags.CANNOT_SUPPORT_KELP))
-            )
-        );
         PlacementUtils.register(
             context,
             KELP_COLD,
@@ -91,7 +_,13 @@
             NoiseBasedCountPlacement.of(120, 80.0, 0.0),
             InSquarePlacement.spread(),
             PlacementUtils.HEIGHTMAP_OCEAN_FLOOR,
-            kelpPlacementFilter,
+            BlockPredicateFilter.forPredicate(
+                BlockPredicate.allOf(
+                    BlockPredicate.matchesBlocks(Blocks.WATER),
+                    BlockPredicate.matchesBlocks(Direction.UP, Blocks.WATER),
+                    BlockPredicate.not(BlockPredicate.matchesTag(BlockTags.CANNOT_SUPPORT_KELP))
+                )
+            ),
             BiomeFilter.biome()
         );
         PlacementUtils.register(
@@ -101,7 +_,13 @@
             NoiseBasedCountPlacement.of(80, 80.0, 0.0),
             InSquarePlacement.spread(),
             PlacementUtils.HEIGHTMAP_OCEAN_FLOOR,
-            kelpPlacementFilter,
+            BlockPredicateFilter.forPredicate(
+                BlockPredicate.allOf(
+                    BlockPredicate.matchesBlocks(Blocks.WATER),
+                    BlockPredicate.matchesBlocks(Direction.UP, Blocks.WATER),
+                    BlockPredicate.not(BlockPredicate.matchesTag(BlockTags.CANNOT_SUPPORT_KELP))
+                )
+            ),
             BiomeFilter.biome()
         );
         PlacementUtils.register(
