--- a/net/minecraft/data/worldgen/placement/MiscOverworldPlacements.java
+++ b/net/minecraft/data/worldgen/placement/MiscOverworldPlacements.java
@@ -4,7 +_,6 @@
 import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
-import net.minecraft.core.Vec3i;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.data.worldgen.BootstrapContext;
 import net.minecraft.data.worldgen.features.MiscOverworldFeatures;
@@ -176,20 +_,7 @@
         PlacementUtils.register(context, FREEZE_TOP_LAYER, freezeTopLayer, BiomeFilter.biome());
         PlacementUtils.register(context, VOID_START_PLATFORM, voidStartPlatform, BiomeFilter.biome());
         PlacementUtils.register(
-            context,
-            DESERT_WELL,
-            desertWell,
-            RarityFilter.onAverageOnceEvery(1000),
-            InSquarePlacement.spread(),
-            PlacementUtils.HEIGHTMAP,
-            OffsetPlacement.of(Direction.DOWN),
-            BlockPredicateFilter.forPredicate(
-                BlockPredicate.allOf(
-                    BlockPredicate.matchesBlocks(Blocks.SAND),
-                    BlockPredicate.volumeMatch(new Vec3i(-2, -2, -2), new Vec3i(2, -1, 2), BlockPredicate.not(BlockPredicate.ONLY_IN_AIR_PREDICATE))
-                )
-            ),
-            BiomeFilter.biome()
+            context, DESERT_WELL, desertWell, RarityFilter.onAverageOnceEvery(1000), InSquarePlacement.spread(), PlacementUtils.HEIGHTMAP, BiomeFilter.biome()
         );
         PlacementUtils.register(
             context,
