--- a/net/minecraft/world/entity/animal/rabbit/Rabbit.java
+++ b/net/minecraft/world/entity/animal/rabbit/Rabbit.java
@@ -599,7 +_,7 @@
         @Override
         public boolean canUse() {
             if (this.nextStartTick <= 0) {
-                if (!getServerLevel(this.rabbit).getGameRules().get(GameRules.MOB_GRIEFING)) {
+                if (!net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(this.rabbit.level()), this.rabbit)) {
                     return false;
                 }
 
@@ -629,9 +_,19 @@
                 if (this.canRaid && block instanceof CarrotBlock) {
                     int carrotAge = blockState.getValue(CarrotBlock.AGE);
                     if (carrotAge == 0) {
+                        // CraftBukkit start
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.rabbit, cropsPos, Blocks.AIR.defaultBlockState())) {
+                            return;
+                        }
+                        // CraftBukkit end
                         level.setBlock(cropsPos, Blocks.AIR.defaultBlockState(), 2);
                         level.destroyBlock(cropsPos, true, this.rabbit);
                     } else {
+                        // CraftBukkit start
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.rabbit, cropsPos, blockState.setValue(CarrotBlock.AGE, carrotAge - 1))) {
+                            return;
+                        }
+                        // CraftBukkit end
                         level.setBlock(cropsPos, blockState.setValue(CarrotBlock.AGE, carrotAge - 1), 2);
                         level.gameEvent(GameEvent.BLOCK_CHANGE, cropsPos, GameEvent.Context.of(this.rabbit));
                         level.levelEvent(2001, cropsPos, Block.getId(blockState));
