--- a/net/minecraft/world/entity/animal/rabbit/Rabbit.java
+++ b/net/minecraft/world/entity/animal/rabbit/Rabbit.java
@@ -108,12 +_,41 @@
     private int jumpDelayTicks;
     public int moreCarrotTicks = 0;
 
+    // Purpur start - Make entity breeding times configurable
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level().purpurConfig.rabbitBreedingTicks;
+    }
+    // Purpur end - Make entity breeding times configurable
+
     public Rabbit(EntityType<? extends Rabbit> type, Level level) {
         super(type, level);
         this.jumpControl = new Rabbit.RabbitJumpControl(this);
         this.moveControl = new Rabbit.RabbitMoveControl<>(this);
-        this.setSpeedModifier(0.0);
-    }
+        // this.setSpeedModifier(0.0); // CraftBukkit
+    }
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.rabbitTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.rabbitMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.rabbitScale);
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.rabbitAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
 
     @Override
     public void registerGoals() {
@@ -417,10 +_,22 @@
         }
 
         this.setVariant(variant);
+
+        // Purpur start - Special mobs naturally spawn
+        if (variant != Variant.EVIL && level.getLevel().purpurConfig.rabbitNaturalToast > 0D && random.nextDouble() <= level.getLevel().purpurConfig.rabbitNaturalToast) {
+            setCustomName(Component.translatable("Toast"));
+        }
+        // Purpur end - Special mobs naturally spawn
         return super.finalizeSpawn(level, difficulty, spawnReason, groupData);
     }
 
     private static Rabbit.Variant getRandomRabbitVariant(LevelAccessor level, BlockPos pos) {
+        // Purpur start - Special mobs naturally spawn
+        Level world = level.getMinecraftWorld();
+        if (world.purpurConfig.rabbitNaturalKiller > 0D && world.getRandom().nextDouble() <= world.purpurConfig.rabbitNaturalKiller) {
+            return Rabbit.Variant.EVIL;
+        }
+        // Purpur end - Special mobs naturally spawn
         Holder<Biome> biome = level.getBiome(pos);
         int randomVal = level.getRandom().nextInt(100);
         if (biome.is(BiomeTags.SPAWNS_WHITE_RABBITS)) {
@@ -432,7 +_,7 @@
         }
     }
 
-    public static boolean checkRabbitSpawnRules(EntityType<Rabbit> type, LevelAccessor level, EntitySpawnReason spawnReason, BlockPos pos, RandomSource random) {
+    public static boolean checkRabbitSpawnRules(EntityType<? extends Rabbit> type, LevelAccessor level, EntitySpawnReason spawnReason, BlockPos pos, RandomSource random) {
         return level.getBlockState(pos.below()).is(BlockTags.RABBITS_SPAWNABLE_ON) && isBrightEnoughToSpawn(level, pos);
     }
 
@@ -599,7 +_,7 @@
         @Override
         public boolean canUse() {
             if (this.nextStartTick <= 0) {
-                if (!getServerLevel(this.rabbit).getGameRules().get(GameRules.MOB_GRIEFING)) {
+                if (!net.neoforged.neoforge.event.EventHooks.canEntityGrief(getServerLevel(this.rabbit.level()), this.rabbit)) {
                     return false;
                 }
 
@@ -629,9 +_,19 @@
                 if (this.canRaid && block instanceof CarrotBlock) {
                     int carrotAge = blockState.getValue(CarrotBlock.AGE);
                     if (carrotAge == 0) {
+                        // CraftBukkit start
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.rabbit, cropsPos, Blocks.AIR.defaultBlockState())) {
+                            return;
+                        }
+                        // CraftBukkit end
                         level.setBlock(cropsPos, Blocks.AIR.defaultBlockState(), 2);
                         level.destroyBlock(cropsPos, true, this.rabbit);
                     } else {
+                        // CraftBukkit start
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.rabbit, cropsPos, blockState.setValue(CarrotBlock.AGE, carrotAge - 1))) {
+                            return;
+                        }
+                        // CraftBukkit end
                         level.setBlock(cropsPos, blockState.setValue(CarrotBlock.AGE, carrotAge - 1), 2);
                         level.gameEvent(GameEvent.BLOCK_CHANGE, cropsPos, GameEvent.Context.of(this.rabbit));
                         level.levelEvent(2001, cropsPos, Block.getId(blockState));
