--- a/com/mojang/realmsclient/gui/screens/configuration/RealmsPlayersTab.java
+++ b/com/mojang/realmsclient/gui/screens/configuration/RealmsPlayersTab.java
@@ -21,7 +_,6 @@
 import net.minecraft.client.gui.components.tabs.GridLayoutTab;
 import net.minecraft.client.gui.layouts.GridLayout;
 import net.minecraft.client.gui.layouts.LayoutSettings;
-import net.minecraft.client.gui.layouts.LinearLayout;
 import net.minecraft.client.gui.narration.NarratableEntry;
 import net.minecraft.client.gui.navigation.ScreenRectangle;
 import net.minecraft.network.chat.CommonComponents;
@@ -54,19 +_,13 @@
             new RealmsPlayersTab.InvitedObjectSelectionList(configurationScreen.width, this.calculateListHeight()),
             LayoutSettings.defaults().alignVerticallyTop().alignHorizontallyCenter()
         );
-        LinearLayout actions = helper.addChild(
-            LinearLayout.horizontal().spacing(8), LayoutSettings.defaults().alignVerticallyBottom().alignHorizontallyCenter()
-        );
-        actions.addChild(
+        helper.addChild(
             Button.builder(
                     Component.translatable("mco.configure.world.buttons.invite"),
                     var3 -> minecraft.gui.setScreen(new RealmsInviteScreen(configurationScreen, serverData))
                 )
-                .build()
-        );
-        actions.addChild(
-            Button.builder(RealmsInviteCodesScreen.TITLE, var3 -> minecraft.gui.setScreen(new RealmsInviteCodesScreen(configurationScreen, serverData)))
-                .build()
+                .build(),
+            LayoutSettings.defaults().alignVerticallyBottom().alignHorizontallyCenter()
         );
         this.updateData(serverData);
     }
