--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -123,6 +123,17 @@
 import net.minecraft.world.scores.Team;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.util.CraftVector;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.entity.EntityExhaustionEvent;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.player.PlayerBedEnterEvent;
+import org.bukkit.event.player.PlayerVelocityEvent;
+// CraftBukkit end
+
 public abstract class Player extends Avatar implements ContainerUser {
 
     public static final int MAX_HEALTH = 20;
@@ -149,7 +160,7 @@
     private static final int DEFAULT_CURRENT_IMPULSE_CONTEXT_RESET_GRACE_TIME = 0;
     public static final float CREATIVE_ENTITY_INTERACTION_RANGE_MODIFIER_VALUE = 2.0F;
     private final Inventory inventory;
-    protected PlayerEnderChestContainer enderChestInventory = new PlayerEnderChestContainer();
+    protected PlayerEnderChestContainer enderChestInventory = new PlayerEnderChestContainer(this); // CraftBukkit - add "this" to constructor
     public final InventoryMenu inventoryMenu;
     public AbstractContainerMenu containerMenu;
     protected FoodData foodData = new FoodData();
@@ -176,6 +187,16 @@
     private boolean ignoreFallDamageFromCurrentImpulse;
     private int currentImpulseContextResetGraceTime;
 
+    // CraftBukkit start
+    public boolean fauxSleeping;
+    public int oldLevel = -1;
+
+    @Override
+    public CraftHumanEntity getBukkitEntity() {
+        return (CraftHumanEntity) super.getBukkitEntity();
+    }
+    // CraftBukkit end
+
     public Player(Level level, GameProfile gameprofile) {
         super(EntityType.PLAYER, level);
         this.lastItemInMainHand = ItemStack.EMPTY;
@@ -322,7 +343,7 @@
     }
 
     private void turtleHelmetTick() {
-        this.addEffect(new MobEffectInstance(MobEffects.WATER_BREATHING, 200, 0, false, false, true));
+        this.addEffect(new MobEffectInstance(MobEffects.WATER_BREATHING, 200, 0, false, false, true), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.TURTLE_HELMET); // CraftBukkit
     }
 
     private boolean isEquipped(Item item) {
@@ -428,8 +449,14 @@
     public void rideTick() {
         if (!this.level().isClientSide() && this.wantsToStopRiding() && this.isPassenger()) {
             this.stopRiding();
-            this.setShiftKeyDown(false);
-        } else {
+            // CraftBukkit start - SPIGOT-7316: no longer passenger, dismount and return
+            if (!this.isPassenger()) {
+                this.setShiftKeyDown(false);
+                return;
+            }
+        }
+        {
+            // CraftBukkit end
             super.rideTick();
         }
     }
@@ -649,7 +676,7 @@
 
         this.setScore(valueinput.getIntOr("Score", 0));
         this.foodData.readAdditionalSaveData(valueinput);
-        Optional optional = valueinput.read("abilities", Abilities.Packed.CODEC);
+        Optional<Abilities.Packed> optional = valueinput.read("abilities", Abilities.Packed.CODEC); // CraftBukkit - decompile error
         Abilities abilities = this.abilities;
 
         Objects.requireNonNull(this.abilities);
@@ -657,7 +684,7 @@
         this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue((double) this.abilities.getWalkingSpeed());
         this.enderChestInventory.fromSlots(valueinput.listOrEmpty("EnderItems", ItemStackWithSlot.CODEC));
         this.setLastDeathLocation(valueinput.read("LastDeathLocation", GlobalPos.CODEC));
-        this.currentImpulseImpactPos = (Vec3) valueinput.read("current_explosion_impact_pos", Vec3.CODEC).orElse((Object) null);
+        this.currentImpulseImpactPos = (Vec3) valueinput.read("current_explosion_impact_pos", Vec3.CODEC).orElse(null); // CraftBukkit - decompile error
         this.ignoreFallDamageFromCurrentImpulse = valueinput.getBooleanOr("ignore_fall_damage_from_current_explosion", false);
         this.currentImpulseContextResetGraceTime = valueinput.getIntOr("current_impulse_context_reset_grace_time", 0);
     }
@@ -701,10 +728,10 @@
             if (this.isDeadOrDying()) {
                 return false;
             } else {
-                this.removeEntitiesOnShoulder();
+                // this.removeEntitiesOnShoulder(); // CraftBukkit - moved down
                 if (damagesource.scalesWithDifficulty()) {
                     if (serverlevel.getDifficulty() == Difficulty.PEACEFUL) {
-                        f = 0.0F;
+                        return false; // CraftBukkit - f = 0.0f -> return false
                     }
 
                     if (serverlevel.getDifficulty() == Difficulty.EASY) {
@@ -716,7 +743,13 @@
                     }
                 }
 
-                return f == 0.0F ? false : super.hurtServer(serverlevel, damagesource, f);
+                // CraftBukkit start - Don't filter out 0 damage
+                boolean damaged = super.hurtServer(serverlevel, damagesource, f);
+                if (damaged) {
+                    this.removeEntitiesOnShoulder();
+                }
+                return damaged;
+                // CraftBukkit end
             }
         }
     }
@@ -740,10 +773,29 @@
     }
 
     public boolean canHarmPlayer(Player player) {
-        Team team = this.getTeam();
-        Team team1 = player.getTeam();
+        // CraftBukkit start - Change to check OTHER player's scoreboard team according to API
+        // To summarize this method's logic, it's "Can parameter hurt this"
+        org.bukkit.scoreboard.Team team;
+        if (player instanceof ServerPlayer) {
+            ServerPlayer thatPlayer = (ServerPlayer) player;
+            team = thatPlayer.getBukkitEntity().getScoreboard().getPlayerTeam(thatPlayer.getBukkitEntity());
+            if (team == null || team.allowFriendlyFire()) {
+                return true;
+            }
+        } else {
+            // This should never be called, but is implemented anyway
+            org.bukkit.OfflinePlayer thisPlayer = player.level().getCraftServer().getOfflinePlayer(player.getScoreboardName());
+            team = player.level().getCraftServer().getScoreboardManager().getMainScoreboard().getPlayerTeam(thisPlayer);
+            if (team == null || team.allowFriendlyFire()) {
+                return true;
+            }
+        }
 
-        return team == null ? true : (!team.isAlliedTo(team1) ? true : team.isAllowFriendlyFire());
+        if (this instanceof ServerPlayer) {
+            return !team.hasPlayer(((ServerPlayer) this).getBukkitEntity());
+        }
+        return !team.hasPlayer(this.level().getCraftServer().getOfflinePlayer(this.getScoreboardName()));
+        // CraftBukkit end
     }
 
     @Override
@@ -757,7 +809,12 @@
     }
 
     @Override
-    protected void actuallyHurt(ServerLevel serverlevel, DamageSource damagesource, float f) {
+    // CraftBukkit start
+    protected boolean actuallyHurt(ServerLevel serverlevel, DamageSource damagesource, float f, EntityDamageEvent event) { // void -> boolean
+        if (true) {
+            return super.actuallyHurt(serverlevel, damagesource, f, event);
+        }
+        // CraftBukkit end
         if (!this.isInvulnerableTo(serverlevel, damagesource)) {
             f = this.getDamageAfterArmorAbsorb(damagesource, f);
             f = this.getDamageAfterMagicAbsorb(damagesource, f);
@@ -772,7 +829,7 @@
             }
 
             if (f != 0.0F) {
-                this.causeFoodExhaustion(damagesource.getFoodExhaustion());
+                this.causeFoodExhaustion(damagesource.getFoodExhaustion(), EntityExhaustionEvent.ExhaustionReason.DAMAGED); // CraftBukkit - EntityExhaustionEvent
                 this.getCombatTracker().recordDamage(damagesource, f);
                 this.setHealth(this.getHealth() - f);
                 if (f < 3.4028235E37F) {
@@ -782,6 +839,7 @@
                 this.gameEvent(GameEvent.ENTITY_DAMAGE);
             }
         }
+        return false; // CraftBukkit
     }
 
     public boolean isTextFilteringEnabled() {
@@ -941,7 +999,7 @@
 
             f *= this.baseDamageScaleFactor();
             this.onAttack();
-            if (!this.deflectProjectile(entity)) {
+            if (!this.deflectProjectile(entity, damagesource, f2)) { // CraftBukkit
                 if (f > 0.0F || f2 > 0.0F) {
                     boolean flag = f1 > 0.9F;
                     boolean flag1;
@@ -986,6 +1044,11 @@
                         this.causeFoodExhaustion(0.1F);
                     } else {
                         this.playServerSideSound(SoundEvents.PLAYER_ATTACK_NODAMAGE);
+                        // CraftBukkit start - resync on cancelled event
+                        if (this instanceof ServerPlayer) {
+                            ((ServerPlayer) this).getBukkitEntity().updateInventory();
+                        }
+                        // CraftBukkit end
                     }
                 }
 
@@ -1008,8 +1071,13 @@
         return !entity.isAttackable() ? true : entity.skipAttackInteraction(this);
     }
 
-    private boolean deflectProjectile(Entity entity) {
+    private boolean deflectProjectile(Entity entity, DamageSource damagesource, float f1) { // CraftBukkit
         if (entity.getType().is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entity instanceof Projectile projectile) {
+            // CraftBukkit start
+            if (CraftEventFactory.handleNonLivingEntityDamageEvent(entity, damagesource, f1, false)) {
+                return true;
+            }
+            // CraftBukkit end
             if (projectile.deflect(ProjectileDeflection.AIM_DEFLECT, this, EntityReference.of(this), true)) {
                 this.level().playSound((Entity) null, this.getX(), this.getY(), this.getZ(), SoundEvents.PLAYER_ATTACK_NODAMAGE, this.getSoundSource());
                 return true;
@@ -1108,7 +1176,7 @@
             if (entity instanceof LivingEntity) {
                 LivingEntity livingentity = (LivingEntity) entity;
 
-                livingentity.knockback((double) f, (double) Mth.sin((double) (this.getYRot() * ((float) Math.PI / 180F))), (double) (-Mth.cos((double) (this.getYRot() * ((float) Math.PI / 180F)))));
+                livingentity.knockback((double) f, (double) Mth.sin((double) (this.getYRot() * ((float) Math.PI / 180F))), (double) (-Mth.cos((double) (this.getYRot() * ((float) Math.PI / 180F)))), this, org.bukkit.event.entity.EntityKnockbackEvent.KnockbackCause.ENTITY_ATTACK); // CraftBukkit
             } else {
                 entity.push((double) (-Mth.sin((double) (this.getYRot() * ((float) Math.PI / 180F))) * f), 0.1D, (double) (Mth.cos((double) (this.getYRot() * ((float) Math.PI / 180F))) * f));
             }
@@ -1118,9 +1186,26 @@
         }
 
         if (entity instanceof ServerPlayer && entity.hurtMarked) {
+            // CraftBukkit start - Add Velocity Event
+            boolean cancelled = false;
+            org.bukkit.entity.Player player = (org.bukkit.entity.Player) entity.getBukkitEntity();
+            org.bukkit.util.Vector velocity = CraftVector.toBukkit(vec3);
+
+            PlayerVelocityEvent event = new PlayerVelocityEvent(player, velocity.clone());
+            this.level().getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                cancelled = true;
+            } else if (!velocity.equals(event.getVelocity())) {
+                player.setVelocity(event.getVelocity());
+            }
+
+            if (!cancelled) {
             ((ServerPlayer) entity).connection.send(new ClientboundSetEntityMotionPacket(entity));
             entity.hurtMarked = false;
             entity.setDeltaMovement(vec3);
+            }
+            // CraftBukkit end
         }
 
     }
@@ -1151,7 +1236,7 @@
                         float f3 = this.getEnchantedDamage(livingentity, f2, damagesource) * f1;
 
                         if (livingentity.hurtServer(serverlevel, damagesource, f3)) {
-                            livingentity.knockback((double) 0.4F, (double) Mth.sin((double) (this.getYRot() * ((float) Math.PI / 180F))), (double) (-Mth.cos((double) (this.getYRot() * ((float) Math.PI / 180F)))));
+                            livingentity.knockback((double) 0.4F, (double) Mth.sin((double) (this.getYRot() * ((float) Math.PI / 180F))), (double) (-Mth.cos((double) (this.getYRot() * ((float) Math.PI / 180F)))), this, org.bukkit.event.entity.EntityKnockbackEvent.KnockbackCause.SWEEP_ATTACK); // CraftBukkit
                             EnchantmentHelper.doPostAttackEffects(serverlevel, livingentity, damagesource);
                         }
                     }
@@ -1196,7 +1281,7 @@
                 f *= this.baseDamageScaleFactor();
             }
 
-            if (flag1 && this.deflectProjectile(entity)) {
+            if (flag1 && this.deflectProjectile(entity, damagesource, f1)) { // CraftBukkit
                 return true;
             } else {
                 float f2 = flag ? f + f1 : 0.0F;
@@ -1229,7 +1314,7 @@
                     this.setLastHurtMob(entity);
                     this.itemAttackInteraction(entity, itemstack, damagesource, flag3);
                     this.damageStatsAndHearts(entity, f3);
-                    this.causeFoodExhaustion(0.1F);
+                    this.causeFoodExhaustion(0.1F, EntityExhaustionEvent.ExhaustionReason.ATTACK); // CraftBukkit - EntityExhaustionEvent
                     return true;
                 }
             }
@@ -1240,7 +1325,14 @@
 
     @Override
     public void remove(Entity.RemovalReason entity_removalreason) {
-        super.remove(entity_removalreason);
+        // CraftBukkit start - add Bukkit remove cause
+        this.remove(entity_removalreason, null);
+    }
+
+    @Override
+    public void remove(Entity.RemovalReason entity_removalreason, EntityRemoveEvent.Cause cause) {
+        super.remove(entity_removalreason, cause);
+        // CraftBukkit end
         this.inventoryMenu.removed(this);
         if (this.hasContainerOpen()) {
             this.doCloseContainer();
@@ -1308,6 +1400,12 @@
     }
 
     public Either<Player.BedSleepingProblem, Unit> startSleepInBed(BlockPos blockpos) {
+        // CraftBukkit start
+        return this.startSleepInBed(blockpos, false);
+    }
+
+    public Either<Player.BedSleepingProblem, Unit> startSleepInBed(BlockPos blockpos, boolean force) {
+        // CraftBukkit end
         this.startSleeping(blockpos);
         this.sleepCounter = 0;
         return Either.right(Unit.INSTANCE);
@@ -1462,7 +1560,15 @@
     }
 
     public void startFallFlying() {
-        this.setSharedFlag(7, true);
+        // CraftBukkit start
+        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callToggleGlideEvent(this, true).isCancelled()) {
+            this.setSharedFlag(7, true);
+        } else {
+            // SPIGOT-5542: must toggle like below
+            this.setSharedFlag(7, true);
+            this.setSharedFlag(7, false);
+        }
+        // CraftBukkit end
     }
 
     @Override
@@ -1577,10 +1683,21 @@
         return this.experienceLevel >= 30 ? 112 + (this.experienceLevel - 30) * 9 : (this.experienceLevel >= 15 ? 37 + (this.experienceLevel - 15) * 5 : 7 + this.experienceLevel * 2);
     }
 
+    // CraftBukkit start
     public void causeFoodExhaustion(float f) {
+        this.causeFoodExhaustion(f, EntityExhaustionEvent.ExhaustionReason.UNKNOWN);
+    }
+
+    public void causeFoodExhaustion(float f, EntityExhaustionEvent.ExhaustionReason reason) {
+        // CraftBukkit end
         if (!this.abilities.invulnerable) {
             if (!this.level().isClientSide()) {
-                this.foodData.addExhaustion(f);
+                // CraftBukkit start
+                EntityExhaustionEvent event = CraftEventFactory.callPlayerExhaustionEvent(this, reason, f);
+                if (!event.isCancelled()) {
+                    this.foodData.addExhaustion(event.getExhaustion());
+                }
+                // CraftBukkit end
             }
 
         }
@@ -1787,7 +1904,7 @@
 
     protected static Optional<Parrot.Variant> extractParrotVariant(CompoundTag compoundtag) {
         if (!compoundtag.isEmpty()) {
-            EntityType<?> entitytype = (EntityType) compoundtag.read("id", EntityType.CODEC).orElse((Object) null);
+            EntityType<?> entitytype = (EntityType) compoundtag.read("id", EntityType.CODEC).orElse(null); // CraftBukkit - decompile error
 
             if (entitytype == EntityType.PARROT) {
                 return compoundtag.read("Variant", Parrot.Variant.LEGACY_CODEC);
@@ -1849,7 +1966,7 @@
 
     @Override
     public void onAttack() {
-        this.resetOnlyAttackStrengthTicker();
+        // this.resetAttackStrengthTicker(); // CraftBukkit - Moved to EntityLiving to reset the cooldown after the damage is dealt
         super.onAttack();
     }
 
@@ -2080,11 +2197,16 @@
         return MoreObjects.toStringHelper(this).add("name", this.getPlainTextName()).add("id", this.getId()).add("pos", this.position()).add("mode", this.gameMode()).add("permission", this.permissions()).toString();
     }
 
-    public static record BedSleepingProblem(@Nullable Component message) {
+    // CraftBukkit start
+    public static record BedSleepingProblem(@Nullable Component message, PlayerBedEnterEvent.BedEnterResult bukkit) {
 
-        public static final Player.BedSleepingProblem TOO_FAR_AWAY = new Player.BedSleepingProblem(Component.translatable("block.minecraft.bed.too_far_away"));
-        public static final Player.BedSleepingProblem OBSTRUCTED = new Player.BedSleepingProblem(Component.translatable("block.minecraft.bed.obstructed"));
-        public static final Player.BedSleepingProblem OTHER_PROBLEM = new Player.BedSleepingProblem((Component) null);
-        public static final Player.BedSleepingProblem NOT_SAFE = new Player.BedSleepingProblem(Component.translatable("block.minecraft.bed.not_safe"));
+        public BedSleepingProblem(@Nullable Component message) {
+            this(message, PlayerBedEnterEvent.BedEnterResult.OTHER_PROBLEM);
+        }
+        public static final Player.BedSleepingProblem TOO_FAR_AWAY = new Player.BedSleepingProblem(Component.translatable("block.minecraft.bed.too_far_away"), PlayerBedEnterEvent.BedEnterResult.TOO_FAR_AWAY);
+        public static final Player.BedSleepingProblem OBSTRUCTED = new Player.BedSleepingProblem(Component.translatable("block.minecraft.bed.obstructed"), PlayerBedEnterEvent.BedEnterResult.OTHER_PROBLEM);
+        public static final Player.BedSleepingProblem OTHER_PROBLEM = new Player.BedSleepingProblem((Component) null, PlayerBedEnterEvent.BedEnterResult.OTHER_PROBLEM);
+        public static final Player.BedSleepingProblem NOT_SAFE = new Player.BedSleepingProblem(Component.translatable("block.minecraft.bed.not_safe"), PlayerBedEnterEvent.BedEnterResult.NOT_SAFE);
+        // CraftBukkit end
     }
 }
