--- a/net/minecraft/world/level/levelgen/flat/FlatLevelGeneratorPreset.java
+++ b/net/minecraft/world/level/levelgen/flat/FlatLevelGeneratorPreset.java
@@ -4,7 +_,7 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
+import net.minecraft.resources.RegistryFileCodec;
 import net.minecraft.world.item.Item;
 
 public record FlatLevelGeneratorPreset(Holder<Item> displayItem, FlatLevelGeneratorSettings settings) {
@@ -14,5 +_,5 @@
             )
             .apply(i, FlatLevelGeneratorPreset::new)
     );
-    public static final Codec<Holder<FlatLevelGeneratorPreset>> CODEC = RegistryCodecs.holder(Registries.FLAT_LEVEL_GENERATOR_PRESET, DIRECT_CODEC);
+    public static final Codec<Holder<FlatLevelGeneratorPreset>> CODEC = RegistryFileCodec.create(Registries.FLAT_LEVEL_GENERATOR_PRESET, DIRECT_CODEC);
 }
