--- a/net/minecraft/client/model/animal/allay/AllayModel.java
+++ b/net/minecraft/client/model/animal/allay/AllayModel.java
@@ -123,7 +_,7 @@
         this.root.translateAndRotate(poseStack);
         this.body.translateAndRotate(poseStack);
         poseStack.translate(0.0F, 0.0625F, 0.1875F);
-        poseStack.rotate(Axis.XP, this.right_arm.xRot);
+        poseStack.mulPose(Axis.XP.rotation(this.right_arm.xRot));
         poseStack.scale(0.7F, 0.7F, 0.7F);
         poseStack.translate(0.0625F, 0.0F, 0.0F);
     }
