--- a/net/minecraft/world/entity/projectile/arrow/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/SpectralArrow.java
@@ -44,7 +44,7 @@
         super.doPostHurtEffects(livingentity);
         MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
 
-        livingentity.addEffect(mobeffectinstance, this.getEffectSource());
+        livingentity.addEffect(mobeffectinstance, this.getEffectSource(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
     }
 
     @Override
