--- a/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
+++ b/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
@@ -27,6 +27,10 @@
 import net.minecraft.world.level.storage.DimensionDataStorage;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import net.minecraft.world.level.dimension.LevelStem;
+// CraftBukkit end
+
 public class LegacyStructureDataHandler implements LegacyTagFixer {
 
     public static final int LAST_MONOLYTH_STRUCTURE_DATA_VERSION = 1493;
@@ -216,7 +220,7 @@
                             Optional<String> optional = listtag.getCompound(0).flatMap((compoundtag2) -> {
                                 return compoundtag2.getString("id");
                             });
-                            Map map = LegacyStructureDataHandler.LEGACY_TO_CURRENT_MAP;
+                            Map<String, String> map = LegacyStructureDataHandler.LEGACY_TO_CURRENT_MAP; // CraftBukkit - decompile error
 
                             Objects.requireNonNull(map);
                             optional.map(map::get).ifPresent((s2) -> {
@@ -253,18 +257,18 @@
         }
     }
 
-    public static Supplier<LegacyTagFixer> getLegacyTagFixer(ResourceKey<Level> resourcekey, Supplier<@Nullable DimensionDataStorage> supplier, DataFixer datafixer) {
-        if (resourcekey == Level.OVERWORLD) {
+    public static Supplier<LegacyTagFixer> getLegacyTagFixer(ResourceKey<LevelStem> resourcekey, Supplier<@Nullable DimensionDataStorage> supplier, DataFixer datafixer) { // CraftBukkit
+        if (resourcekey == LevelStem.OVERWORLD) { // CraftBukkit
             return () -> {
                 return new LegacyStructureDataHandler((DimensionDataStorage) supplier.get(), ImmutableList.of("Monument", "Stronghold", "Village", "Mineshaft", "Temple", "Mansion"), ImmutableList.of("Village", "Mineshaft", "Mansion", "Igloo", "Desert_Pyramid", "Jungle_Pyramid", "Swamp_Hut", "Stronghold", "Monument"), datafixer);
             };
-        } else if (resourcekey == Level.NETHER) {
+        } else if (resourcekey == LevelStem.NETHER) { // CraftBukkit
             List<String> list = ImmutableList.of("Fortress");
 
             return () -> {
                 return new LegacyStructureDataHandler((DimensionDataStorage) supplier.get(), list, list, datafixer);
             };
-        } else if (resourcekey == Level.END) {
+        } else if (resourcekey == LevelStem.END) { // CraftBukkit
             List<String> list1 = ImmutableList.of("EndCity");
 
             return () -> {
