--- a/net/minecraft/world/entity/monster/cubemob/SulfurCube.java
+++ b/net/minecraft/world/entity/monster/cubemob/SulfurCube.java
@@ -75,6 +_,7 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityRemoveEvent;
 import org.jspecify.annotations.Nullable;
 
 public class SulfurCube extends AbstractCubeMob implements Bucketable, Shearable {
@@ -318,6 +_,7 @@
                     this.triggerOnDeathMobEffects(level, Entity.RemovalReason.KILLED);
                 }
 
+                this.removeReason(EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit
                 this.discard();
             }
         }
@@ -556,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource source) {
+    public SoundEvent getHurtSound(DamageSource source) {
         return this.isTiny() ? SoundEvents.SULFUR_CUBE_SMALL_HURT : SoundEvents.SULFUR_CUBE_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return this.isTiny() ? SoundEvents.SULFUR_CUBE_SMALL_DEATH : SoundEvents.SULFUR_CUBE_DEATH;
     }
 
