--- a/net/minecraft/world/entity/monster/cubemob/SulfurCube.java
+++ b/net/minecraft/world/entity/monster/cubemob/SulfurCube.java
@@ -75,6 +_,7 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityRemoveEvent;
 import org.jspecify.annotations.Nullable;
 
 public class SulfurCube extends AbstractCubeMob implements Bucketable, Shearable {
@@ -95,7 +_,7 @@
     private SulfurCubeArchetype.SoundSettings soundSettings = SulfurCubeArchetype.DEFAULT_SOUND_SETTINGS;
     private int fuse = -1;
     private List<SulfurCubeArchetype.ContactDamage> contactDamages = new ArrayList<>();
-    private static final EntityDataAccessor<Integer> MAX_FUSE = SynchedEntityData.defineId(SulfurCube.class, EntityDataSerializers.INT);
+    public static final EntityDataAccessor<Integer> MAX_FUSE = SynchedEntityData.defineId(SulfurCube.class, EntityDataSerializers.INT);
     private static final EntityDataAccessor<Boolean> FROM_BUCKET = SynchedEntityData.defineId(SulfurCube.class, EntityDataSerializers.BOOLEAN);
     private static final boolean DEFAULT_FROM_BUCKET = false;
     private static final float HORIZONTAL_HIT_ANGLE_SCALE = 1.6F;
@@ -142,7 +_,7 @@
         return this.getFuse() >= 0;
     }
 
-    private void setFuse(int fuse) {
+    public void setFuse(int fuse) {
         this.fuse = fuse;
     }
 
@@ -318,7 +_,7 @@
                     this.triggerOnDeathMobEffects(level, Entity.RemovalReason.KILLED);
                 }
 
-                this.discard();
+                this.discard(EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit
             }
         }
     }
@@ -556,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource source) {
+    public SoundEvent getHurtSound(DamageSource source) {
         return this.isTiny() ? SoundEvents.SULFUR_CUBE_SMALL_HURT : SoundEvents.SULFUR_CUBE_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return this.isTiny() ? SoundEvents.SULFUR_CUBE_SMALL_DEATH : SoundEvents.SULFUR_CUBE_DEATH;
     }
 
