--- a/net/minecraft/world/flag/FeatureFlag.java
+++ b/net/minecraft/world/flag/FeatureFlag.java
@@ -1,8 +_,8 @@
 package net.minecraft.world.flag;
 
 public class FeatureFlag {
-    public final FeatureFlagUniverse universe;
-    public final long mask;
+    final FeatureFlagUniverse universe;
+    final long mask;
 
     FeatureFlag(FeatureFlagUniverse p_249115_, int p_251067_) {
         this.universe = p_249115_;
