--- a/net/minecraft/server/network/config/PrepareSpawnTask.java
+++ b/net/minecraft/server/network/config/PrepareSpawnTask.java
@@ -36,10 +_,18 @@
     private final LevelLoadListener loadListener;
     private PrepareSpawnTask.@Nullable State state;
 
-    public PrepareSpawnTask(MinecraftServer server, NameAndId nameAndId) {
+    // Paper start - passthrough profile and packet listener
+    private final com.mojang.authlib.GameProfile profile;
+    private final net.minecraft.server.network.ServerConfigurationPacketListenerImpl listener;
+    private boolean newPlayer;
+
+    public PrepareSpawnTask(MinecraftServer server, final com.mojang.authlib.GameProfile profile, final net.minecraft.server.network.ServerConfigurationPacketListenerImpl listener) {
+        this.profile = profile;
+        this.listener = listener;
+        // Paper end - passthrough profile and packet listener
         this.server = server;
-        this.nameAndId = nameAndId;
-        this.loadListener = server.getLevelLoadListener();
+        this.nameAndId = new net.minecraft.server.players.NameAndId(profile); // Paper - passthrough profile and packet listener - create from profile
+        this.loadListener = LevelLoadListener.noop(); // Paper - per level load listener - this is already a no-op on dedicated server, but we moved it to Level
     }
 
     @Override
@@ -110,11 +_,12 @@
     }
 
     private final class Preparing implements PrepareSpawnTask.State {
-        private final ServerLevel spawnLevel;
-        private final CompletableFuture<Vec3> spawnPosition;
-        private final Vec2 spawnAngle;
+        private ServerLevel spawnLevel; // Paper - remove final
+        private CompletableFuture<Vec3> spawnPosition; // Paper - remove final
+        private Vec2 spawnAngle; // Paper - remove final
         private @Nullable CompletableFuture<?> chunkLoadFuture;
-        private final ChunkLoadCounter chunkLoadCounter = new ChunkLoadCounter();
+        private @Nullable CompletableFuture<org.bukkit.Location> eventFuture; // Paper
+        private final ChunkLoadCounter chunkLoadCounter = new ca.spottedleaf.moonrise.patches.chunk_system.MoonriseChunkLoadCounter(); // Paper - rewrite chunk system
 
         private Preparing(ServerLevel spawnLevel, CompletableFuture<Vec3> spawnPosition, Vec2 spawnAngle) {
             this.spawnLevel = spawnLevel;
@@ -133,12 +_,56 @@
 
             Vec3 spawnPosition = this.spawnPosition.join();
             if (this.chunkLoadFuture == null) {
-                ChunkPos spawnChunk = ChunkPos.containing(BlockPos.containing(spawnPosition));
-                this.chunkLoadCounter
-                    .track(
-                        this.spawnLevel,
-                        () -> this.chunkLoadFuture = this.spawnLevel.getChunkSource().addTicketAndLoadWithRadius(TicketType.PLAYER_SPAWN, spawnChunk, 3)
+                // Paper start - PlayerSpawnLocationEvent
+                if (this.eventFuture == null && org.spigotmc.event.player.PlayerSpawnLocationEvent.getHandlerList().getRegisteredListeners().length != 0) {
+                    ServerPlayer player;
+                    if (PrepareSpawnTask.this.listener.connection.savedPlayerForLegacyEvents != null) {
+                        player = PrepareSpawnTask.this.listener.connection.savedPlayerForLegacyEvents;
+                    } else {
+                        player = new ServerPlayer(
+                                PrepareSpawnTask.this.server,
+                                PrepareSpawnTask.this.server.overworld(),
+                                PrepareSpawnTask.this.profile,
+                                net.minecraft.server.level.ClientInformation.createDefault()
+                        );
+                        PrepareSpawnTask.this.listener.connection.savedPlayerForLegacyEvents = player;
+                    }
+                    org.spigotmc.event.player.PlayerSpawnLocationEvent ev = new org.spigotmc.event.player.PlayerSpawnLocationEvent(
+                            player.getBukkitEntity(),
+                            org.bukkit.craftbukkit.util.CraftLocation.toBukkit(spawnPosition, this.spawnLevel, this.spawnAngle.x, this.spawnAngle.y)
                     );
+                    ev.callEvent();
+                    spawnPosition = io.papermc.paper.util.MCUtil.toVec3(ev.getSpawnLocation());
+                    if (ev.getSpawnLocation().getWorld() != null) this.spawnLevel = ((org.bukkit.craftbukkit.CraftWorld) ev.getSpawnLocation().getWorld()).getHandle();
+                    this.spawnPosition = CompletableFuture.completedFuture(spawnPosition);
+                    this.spawnAngle = new Vec2(ev.getSpawnLocation().getYaw(), ev.getSpawnLocation().getPitch());
+                }
+
+                if (this.eventFuture == null && io.papermc.paper.event.player.AsyncPlayerSpawnLocationEvent.getHandlerList().getRegisteredListeners().length != 0) {
+                    final Vec3 spawnPositionFinal = spawnPosition;
+                    this.eventFuture = CompletableFuture.supplyAsync(() -> {
+                        io.papermc.paper.event.player.AsyncPlayerSpawnLocationEvent ev = new io.papermc.paper.event.player.AsyncPlayerSpawnLocationEvent(
+                                PrepareSpawnTask.this.listener.paperConnection,
+                                org.bukkit.craftbukkit.util.CraftLocation.toBukkit(spawnPositionFinal, this.spawnLevel, this.spawnAngle.x, this.spawnAngle.y),
+                                PrepareSpawnTask.this.newPlayer
+                        );
+                        ev.callEvent();
+                        return ev.getSpawnLocation();
+                    }, io.papermc.paper.connection.PaperConfigurationTask.CONFIGURATION_POOL);
+                }
+                if (this.eventFuture != null) {
+                    if (!this.eventFuture.isDone()) {
+                        return null;
+                    }
+                    org.bukkit.Location location = this.eventFuture.join();
+                    spawnPosition = io.papermc.paper.util.MCUtil.toVec3(location);
+                    this.spawnLevel = ((org.bukkit.craftbukkit.CraftWorld) location.getWorld()).getHandle();
+                    this.spawnPosition = CompletableFuture.completedFuture(spawnPosition);
+                    this.spawnAngle = new Vec2(location.getYaw(), location.getPitch());
+                }
+                // Paper end - PlayerSpawnLocationEvent
+                ChunkPos spawnChunk = ChunkPos.containing(BlockPos.containing(spawnPosition));
+                this.chunkLoadFuture = ((ca.spottedleaf.moonrise.patches.chunk_system.MoonriseChunkLoadCounter)this.chunkLoadCounter).trackLoadWithRadius(this.spawnLevel, spawnChunk, 3, net.minecraft.world.level.chunk.status.ChunkStatus.FULL, ca.spottedleaf.concurrentutil.util.Priority.HIGH, () -> { Preparing.this.spawnLevel.getChunkSource().addTicketWithRadius(TicketType.PLAYER_SPAWN, spawnChunk, 3); }); // Paper - rewrite chunk system
                 PrepareSpawnTask.this.loadListener.start(LevelLoadListener.Stage.LOAD_PLAYER_CHUNKS, this.chunkLoadCounter.totalChunks());
                 PrepareSpawnTask.this.loadListener.updateFocus(this.spawnLevel.dimension(), spawnChunk);
             }
@@ -172,7 +_,19 @@
         public ServerPlayer spawn(Connection connection, CommonListenerCookie cookie) {
             ChunkPos spawnChunk = ChunkPos.containing(BlockPos.containing(this.spawnPosition));
             this.spawnLevel.waitForEntities(spawnChunk, 3);
-            ServerPlayer player = new ServerPlayer(PrepareSpawnTask.this.server, this.spawnLevel, cookie.gameProfile(), cookie.clientInformation());
+            // Paper start - configuration api - possibly use legacy saved server player instance
+            ServerPlayer player;
+            if (connection.savedPlayerForLegacyEvents != null) {
+                player = connection.savedPlayerForLegacyEvents;
+                connection.savedPlayerForLegacyEvents = null;
+                // Update the existing instance
+                player.gameProfile = cookie.gameProfile();
+                player.updateOptionsNoEvents(cookie.clientInformation());
+                player.setServerLevel(this.spawnLevel);
+            } else {
+                player = new ServerPlayer(PrepareSpawnTask.this.server, this.spawnLevel, cookie.gameProfile(), cookie.clientInformation());
+            }
+            // Paper end - configuration api - possibly use legacy saved server player instance
 
             try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(player.problemPath(), PrepareSpawnTask.LOGGER)) {
                 Optional<ValueInput> input = PrepareSpawnTask.this.server
@@ -180,6 +_,17 @@
                     .loadPlayerData(PrepareSpawnTask.this.nameAndId)
                     .map(tag -> TagValueInput.create(reporter, PrepareSpawnTask.this.server.registryAccess(), tag));
                 input.ifPresent(player::load);
+                net.neoforged.neoforge.event.EventHooks.firePlayerLoadingEvent(player, PrepareSpawnTask.this.server.getPlayerList(), player.getStringUUID());
+                // CraftBukkit start - Better rename detection
+                if (input.isPresent()) {
+                    player.lastKnownName = input.flatMap(t -> t.child("bukkit")).flatMap(t -> t.getString("lastKnownName")).orElse(null);
+                }
+                // CraftBukkit end - Better rename detection
+                // Paper start - Entity#getEntitySpawnReason
+                if (input.isEmpty()) {
+                    player.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT; // set Player SpawnReason to DEFAULT on first login
+                }
+                // Paper end - Entity#getEntitySpawnReason
                 player.snapTo(this.spawnPosition, this.spawnAngle.x, this.spawnAngle.y);
                 PrepareSpawnTask.this.server.getPlayerList().placeNewPlayer(connection, player, cookie);
                 input.ifPresent(tag -> {
