--- a/net/minecraft/server/network/config/PrepareSpawnTask.java
+++ b/net/minecraft/server/network/config/PrepareSpawnTask.java
@@ -16,7 +_,6 @@
 import net.minecraft.server.level.progress.LevelLoadListener;
 import net.minecraft.server.network.CommonListenerCookie;
 import net.minecraft.server.network.ConfigurationTask;
-import net.minecraft.server.players.NameAndId;
 import net.minecraft.util.ProblemReporter;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.storage.LevelData;
@@ -32,11 +_,11 @@
     public static final ConfigurationTask.Type TYPE = new ConfigurationTask.Type("prepare_spawn");
     public static final int PREPARE_CHUNK_RADIUS = 3;
     private final MinecraftServer server;
-    private final NameAndId nameAndId;
+    private final ServerPlayer nameAndId;
     private final LevelLoadListener loadListener;
     private PrepareSpawnTask.@Nullable State state;
 
-    public PrepareSpawnTask(MinecraftServer server, NameAndId nameAndId) {
+    public PrepareSpawnTask(MinecraftServer server, ServerPlayer nameAndId) {
         this.server = server;
         this.nameAndId = nameAndId;
         this.loadListener = server.getLevelLoadListener();
@@ -172,7 +_,7 @@
         public ServerPlayer spawn(Connection connection, CommonListenerCookie cookie) {
             ChunkPos spawnChunk = ChunkPos.containing(BlockPos.containing(this.spawnPosition));
             this.spawnLevel.waitForEntities(spawnChunk, 3);
-            ServerPlayer player = new ServerPlayer(PrepareSpawnTask.this.server, this.spawnLevel, cookie.gameProfile(), cookie.clientInformation());
+            ServerPlayer player = nameAndId; // CraftBukkit
 
             try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(player.problemPath(), PrepareSpawnTask.LOGGER)) {
                 Optional<ValueInput> input = PrepareSpawnTask.this.server
@@ -180,6 +_,17 @@
                     .loadPlayerData(PrepareSpawnTask.this.nameAndId)
                     .map(tag -> TagValueInput.create(reporter, PrepareSpawnTask.this.server.registryAccess(), tag));
                 input.ifPresent(player::load);
+                net.neoforged.neoforge.event.EventHooks.firePlayerLoadingEvent(player, PrepareSpawnTask.this.server.getPlayerList(), player.getStringUUID());
+                // CraftBukkit start - Better rename detection
+                if (input.isPresent()) {
+                    player.lastKnownName = input.flatMap(t -> t.child("bukkit")).flatMap(t -> t.getString("lastKnownName")).orElse(null);
+                }
+                // CraftBukkit end - Better rename detection
+                // Paper start - Entity#getEntitySpawnReason
+                if (input.isEmpty()) {
+                    player.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT; // set Player SpawnReason to DEFAULT on first login
+                }
+                // Paper end - Entity#getEntitySpawnReason
                 player.snapTo(this.spawnPosition, this.spawnAngle.x, this.spawnAngle.y);
                 PrepareSpawnTask.this.server.getPlayerList().placeNewPlayer(connection, player, cookie);
                 input.ifPresent(tag -> {
