--- a/net/minecraft/world/entity/animal/AbstractGolem.java
+++ b/net/minecraft/world/entity/animal/AbstractGolem.java
@@ -14,19 +_,19 @@
 
     @Nullable
     @Override
-    protected SoundEvent getAmbientSound() {
-        return null;
-    }
-
-    @Nullable
-    @Override
-    protected SoundEvent getHurtSound(DamageSource p_27517_) {
-        return null;
-    }
-
-    @Nullable
-    @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getAmbientSound() {
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public SoundEvent getHurtSound(DamageSource p_27517_) {
+        return null;
+    }
+
+    @Nullable
+    @Override
+    public SoundEvent getDeathSound() {
         return null;
     }
 
