--- a/net/minecraft/world/entity/animal/Cat.java
+++ b/net/minecraft/world/entity/animal/Cat.java
@@ -220,7 +_,7 @@
 
     @Nullable
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         if (this.isTame()) {
             if (this.isInLove()) {
                 return SoundEvents.CAT_PURR;
@@ -242,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_28160_) {
+    public SoundEvent getHurtSound(DamageSource p_28160_) {
         return SoundEvents.CAT_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.CAT_DEATH;
     }
 
@@ -387,9 +_,9 @@
                     }
                 } else if (this.isFood(itemstack) && this.getHealth() < this.getMaxHealth()) {
                     if (!this.level().isClientSide()) {
-                        this.usePlayerItem(p_28153_, p_28154_, itemstack);
                         FoodProperties foodproperties = itemstack.get(DataComponents.FOOD);
                         this.heal(foodproperties != null ? foodproperties.nutrition() : 1.0F);
+                        this.usePlayerItem(p_28153_, p_28154_, itemstack);
                         this.playEatingSound();
                     }
 
@@ -451,7 +_,7 @@
     }
 
     private void tryToTame(Player p_333858_) {
-        if (this.random.nextInt(3) == 0) {
+        if (this.random.nextInt(3) == 0 && !net.neoforged.neoforge.event.EventHooks.onAnimalTame(this, p_333858_)) {
             this.tame(p_333858_);
             this.setOrderedToSit(true);
             this.level().broadcastEntityEvent(this, (byte)7);
