--- a/net/minecraft/world/entity/animal/IronGolem.java
+++ b/net/minecraft/world/entity/animal/IronGolem.java
@@ -251,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_28872_) {
+    public SoundEvent getHurtSound(DamageSource p_28872_) {
         return SoundEvents.IRON_GOLEM_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.IRON_GOLEM_DEATH;
     }
 
