--- a/net/minecraft/world/entity/animal/Ocelot.java
+++ b/net/minecraft/world/entity/animal/Ocelot.java
@@ -136,7 +_,7 @@
 
     @Nullable
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.OCELOT_AMBIENT;
     }
 
@@ -146,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_29035_) {
+    public SoundEvent getHurtSound(DamageSource p_29035_) {
         return SoundEvents.OCELOT_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.OCELOT_DEATH;
     }
 
@@ -161,7 +_,7 @@
         if ((this.temptGoal == null || this.temptGoal.isRunning()) && !this.isTrusting() && this.isFood(itemstack) && p_29021_.distanceToSqr(this) < 9.0) {
             this.usePlayerItem(p_29021_, p_29022_, itemstack);
             if (!this.level().isClientSide()) {
-                if (this.random.nextInt(3) == 0) {
+                if (this.random.nextInt(3) == 0 && !net.neoforged.neoforge.event.EventHooks.onAnimalTame(this, p_29021_)) {
                     this.setTrusting(true);
                     this.spawnTrustingParticles(true);
                     this.level().broadcastEntityEvent(this, (byte)41);
