--- a/net/minecraft/world/entity/animal/Panda.java
+++ b/net/minecraft/world/entity/animal/Panda.java
@@ -652,7 +_,7 @@
 
     @Nullable
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         if (this.isAggressive()) {
             return SoundEvents.PANDA_AGGRESSIVE_AMBIENT;
         } else {
@@ -672,13 +_,13 @@
 
     @Nullable
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.PANDA_DEATH;
     }
 
     @Nullable
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_29142_) {
+    public SoundEvent getHurtSound(DamageSource p_29142_) {
         return SoundEvents.PANDA_HURT;
     }
 
