--- a/net/minecraft/world/entity/animal/Panda.java
+++ b/net/minecraft/world/entity/animal/Panda.java
@@ -70,6 +_,7 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class Panda extends Animal {
     private static final EntityDataAccessor<Integer> UNHAPPY_COUNTER = SynchedEntityData.defineId(Panda.class, EntityDataSerializers.INT);
@@ -137,6 +_,7 @@
     }
 
     public void sit(boolean p_29209_) {
+        if (!new io.papermc.paper.event.entity.EntityToggleSitEvent(this.getBukkitEntity(), p_29209_).callEvent()) return; // Paper - Add EntityToggleSitEvent
         this.setFlag(8, p_29209_);
     }
 
@@ -528,7 +_,9 @@
 
         for (Panda panda : level.getEntitiesOfClass(Panda.class, this.getBoundingBox().inflate(10.0))) {
             if (!panda.isBaby() && panda.onGround() && !panda.isInWater() && panda.canPerformAction()) {
-                panda.jumpFromGround();
+                if (new com.destroystokyo.paper.event.entity.EntityJumpEvent(getBukkitLivingEntity()).callEvent()) { // Paper - Entity Jump API
+                    panda.jumpFromGround();
+                } else { this.setJumping(false); } // Paper - Entity Jump API; setJumping(false) stops a potential loop
             }
         }
 
@@ -548,7 +_,7 @@
 
     @Override
     protected void pickUpItem(ItemEntity p_29121_) {
-        if (this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && PANDA_ITEMS.test(p_29121_)) {
+        if (!CraftEventFactory.callEntityPickupItemEvent(this, p_29121_, 0, !(this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && Panda.PANDA_ITEMS.test(p_29121_))).isCancelled()) { // CraftBukkit
             this.onItemPickup(p_29121_);
             ItemStack itemstack = p_29121_.getItem();
             this.setItemSlot(EquipmentSlot.MAINHAND, itemstack);
@@ -696,7 +_,7 @@
 
     @Nullable
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.PANDA_DEATH;
     }
 
@@ -883,6 +_,7 @@
         @Override
         protected void alertOther(Mob p_29295_, LivingEntity p_29296_) {
             if (p_29295_ instanceof Panda && p_29295_.isAggressive()) {
+                p_29295_.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit;
                 p_29295_.setTarget(p_29296_);
             }
         }
