--- a/net/minecraft/world/entity/animal/Salmon.java
+++ b/net/minecraft/world/entity/animal/Salmon.java
@@ -50,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.SALMON_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.SALMON_DEATH;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_29795_) {
+    public SoundEvent getHurtSound(DamageSource p_29795_) {
         return SoundEvents.SALMON_HURT;
     }
 
