--- a/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/SnowGolem.java
@@ -32,7 +_,6 @@
 import net.minecraft.world.entity.projectile.Snowball;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
-import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
@@ -95,7 +_,7 @@
                 this.hurtServer(serverlevel, this.damageSources().onFire(), 1.0F);
             }
 
-            if (!serverlevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            if (!net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverlevel, this)) {
                 return;
             }
 
@@ -136,7 +_,7 @@
     @Override
     protected InteractionResult mobInteract(Player p_29920_, InteractionHand p_29921_) {
         ItemStack itemstack = p_29920_.getItemInHand(p_29921_);
-        if (itemstack.is(Items.SHEARS) && this.readyForShearing()) {
+        if (false && itemstack.is(Items.SHEARS) && this.readyForShearing()) { // Neo: Shear logic is handled by IShearable
             if (this.level() instanceof ServerLevel serverlevel) {
                 this.shear(serverlevel, SoundSource.PLAYERS, itemstack);
                 this.gameEvent(GameEvent.SHEAR, p_29920_);
@@ -178,19 +_,19 @@
 
     @Nullable
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.SNOW_GOLEM_AMBIENT;
     }
 
     @Nullable
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_29929_) {
+    public SoundEvent getHurtSound(DamageSource p_29929_) {
         return SoundEvents.SNOW_GOLEM_HURT;
     }
 
     @Nullable
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.SNOW_GOLEM_DEATH;
     }
 
