--- a/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/net/minecraft/world/entity/animal/SnowGolem.java
@@ -31,21 +_,32 @@
 import net.minecraft.world.entity.projectile.Snowball;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
-import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackMob {
     private static final EntityDataAccessor<Byte> DATA_PUMPKIN_ID = SynchedEntityData.defineId(SnowGolem.class, EntityDataSerializers.BYTE);
     private static final byte PUMPKIN_FLAG = 16;
+    @Nullable private java.util.UUID summoner; // Purpur
 
     public SnowGolem(EntityType<? extends SnowGolem> p_29902_, Level p_29903_) {
         super(p_29902_, p_29903_);
     }
 
+
+    @Nullable
+    public java.util.UUID getSummoner() {
+        return summoner;
+    }
+
+    public void setSummoner(@Nullable java.util.UUID summoner) {
+        this.summoner = summoner;
+    }
+
     @Override
     protected void registerGoals() {
         this.goalSelector.addGoal(1, new RangedAttackGoal(this, 1.25, 20, 10.0F));
@@ -69,6 +_,7 @@
     public void addAdditionalSaveData(CompoundTag p_29923_) {
         super.addAdditionalSaveData(p_29923_);
         p_29923_.putBoolean("Pumpkin", this.hasPumpkin());
+        if (getSummoner() != null) p_29923_.putUUID("Purpur.Summoner", getSummoner()); // Purpur
     }
 
     @Override
@@ -77,6 +_,7 @@
         if (p_29915_.contains("Pumpkin")) {
             this.setPumpkin(p_29915_.getBoolean("Pumpkin"));
         }
+        if (p_29915_.contains("Purpur.Summoner")) setSummoner(p_29915_.getUUID("Purpur.Summoner")); // Purpur
     }
 
     @Override
@@ -89,10 +_,10 @@
         super.aiStep();
         if (!this.level().isClientSide) {
             if (this.level().getBiome(this.blockPosition()).is(BiomeTags.SNOW_GOLEM_MELTS)) {
-                this.hurt(this.damageSources().onFire(), 1.0F);
+                this.hurt(this.damageSources().melting(), 1.0F); // CraftBukkit - DamageSources.ON_FIRE -> CraftEventFactory.MELTING
             }
 
-            if (!this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            if (!net.neoforged.neoforge.event.EventHooks.canEntityGrief(this.level(), this)) {
                 return;
             }
 
@@ -104,7 +_,11 @@
                 int l = Mth.floor(this.getZ() + (double)((float)(i / 2 % 2 * 2 - 1) * 0.25F));
                 BlockPos blockpos = new BlockPos(j, k, l);
                 if (this.level().getBlockState(blockpos).isAir() && blockstate.canSurvive(this.level(), blockpos)) {
-                    this.level().setBlockAndUpdate(blockpos, blockstate);
+                    // CraftBukkit start
+                    if (!CraftEventFactory.handleBlockFormEvent(this.level(), blockpos, blockstate, this)) {
+                        continue;
+                    }
+                    // CraftBukkit end
                     this.level().gameEvent(GameEvent.BLOCK_PLACE, blockpos, GameEvent.Context.of(this, blockstate));
                 }
             }
@@ -127,7 +_,7 @@
     @Override
     protected InteractionResult mobInteract(Player p_29920_, InteractionHand p_29921_) {
         ItemStack itemstack = p_29920_.getItemInHand(p_29921_);
-        if (itemstack.is(Items.SHEARS) && this.readyForShearing()) {
+        if (false && itemstack.is(Items.SHEARS) && this.readyForShearing()) { // Neo: Shear logic is handled by IShearable
             this.shear(SoundSource.PLAYERS);
             this.gameEvent(GameEvent.SHEAR, p_29920_);
             if (!this.level().isClientSide) {
@@ -141,12 +_,31 @@
     }
 
     @Override
+    public java.util.List<ItemStack> generateDefaultDrops() {
+        return java.util.Collections.singletonList(new ItemStack(Items.CARVED_PUMPKIN));
+    }
+
+    @Override
     public void shear(SoundSource p_29907_) {
         this.level().playSound(null, this, SoundEvents.SNOW_GOLEM_SHEAR, p_29907_, 1.0F, 1.0F);
         if (!this.level().isClientSide()) {
             this.setPumpkin(false);
-            this.spawnAtLocation(new ItemStack(Items.CARVED_PUMPKIN), this.getEyeHeight());
+            // Paper start - custom shear drops (moved drop generation to separate method)
+            for (final ItemStack drop : shear$drops.isEmpty() ? generateDefaultDrops() : shear$drops) {
+                this.forceDrops = true;
+                this.spawnAtLocation(drop, this.getEyeHeight());
+                this.forceDrops = false;
+            }
+            // Paper end - custom shear drops
         }
+    }
+
+    public java.util.List<net.minecraft.world.item.ItemStack> shear$drops = new java.util.ArrayList<>();
+
+    @Override
+    public void shearPaper(SoundSource p_331493_, java.util.List<net.minecraft.world.item.ItemStack> drops) {
+        shear$drops = drops;
+        this.shear(p_331493_);
     }
 
     @Override
@@ -181,7 +_,7 @@
 
     @Nullable
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.SNOW_GOLEM_DEATH;
     }
 
