--- a/net/minecraft/world/entity/animal/Squid.java
+++ b/net/minecraft/world/entity/animal/Squid.java
@@ -61,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.SQUID_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_29980_) {
+    public SoundEvent getHurtSound(DamageSource p_29980_) {
         return SoundEvents.SQUID_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.SQUID_DEATH;
     }
 
@@ -85,7 +_,7 @@
     }
 
     @Override
-    protected float getSoundVolume() {
+    public float getSoundVolume() {
         return 0.4F;
     }
 
