--- a/net/minecraft/world/entity/animal/Squid.java
+++ b/net/minecraft/world/entity/animal/Squid.java
@@ -41,7 +_,7 @@
 
     public Squid(EntityType<? extends Squid> p_29953_, Level p_29954_) {
         super(p_29953_, p_29954_);
-        this.random.setSeed((long)this.getId());
+        // this.random.setSeed((long)this.getId()); // Paper - Share random for entities to make them more random
         this.tentacleSpeed = 1.0F / (this.random.nextFloat() + 1.0F) * 0.2F;
     }
 
@@ -66,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.SQUID_DEATH;
     }
 
@@ -80,7 +_,7 @@
     }
 
     @Override
-    protected float getSoundVolume() {
+    public float getSoundVolume() {
         return 0.4F;
     }
 
