--- a/net/minecraft/world/entity/animal/TropicalFish.java
+++ b/net/minecraft/world/entity/animal/TropicalFish.java
@@ -214,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.TROPICAL_FISH_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.TROPICAL_FISH_DEATH;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_30039_) {
+    public SoundEvent getHurtSound(DamageSource p_30039_) {
         return SoundEvents.TROPICAL_FISH_HURT;
     }
 
