--- a/net/minecraft/world/entity/animal/WaterAnimal.java
+++ b/net/minecraft/world/entity/animal/WaterAnimal.java
@@ -67,6 +_,10 @@
     ) {
         int i = p_218284_.getSeaLevel();
         int j = i - 13;
+        // Paper start - Make water animal spawn height configurable
+        i = p_218284_.getMinecraftWorld().paperConfig().entities.spawning.wateranimalSpawnHeight.maximum.or(i);
+        j = p_218284_.getMinecraftWorld().paperConfig().entities.spawning.wateranimalSpawnHeight.minimum.or(j);
+        // Paper end - Make water animal spawn height configurable
         return p_218286_.getY() >= j
             && p_218286_.getY() <= i
             && p_218284_.getFluidState(p_218286_.below()).is(FluidTags.WATER)
