--- a/net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/trunkplacers/TrunkPlacer.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.levelgen.feature.TreeFeature;
+import net.minecraft.world.level.levelgen.feature.configurations.TreeConfiguration;
 import net.minecraft.world.level.levelgen.feature.foliageplacers.FoliagePlacer;
 
 public abstract class TrunkPlacer {
@@ -47,7 +_,7 @@
         final RandomSource random,
         final int treeHeight,
         final BlockPos origin,
-        final TreeFeature tree
+        final TreeConfiguration config
     );
 
     public int getBaseHeight() {
@@ -59,16 +_,20 @@
     }
 
     protected static void placeBelowTrunkBlock(
-        WorldGenLevel level, BiConsumer<BlockPos, BlockState> trunkSetter, RandomSource random, BlockPos pos, TreeFeature tree
+        WorldGenLevel level, BiConsumer<BlockPos, BlockState> trunkSetter, RandomSource random, BlockPos pos, TreeConfiguration config
     ) {
-        BlockState blockBelowTrunk = tree.belowTrunkProvider().getOptionalState(level, random, pos);
+        if (level.getBlockState(pos).onTreeGrow(level, trunkSetter, random, pos, config)) {
+            return;
+        }
+
+        BlockState blockBelowTrunk = config.belowTrunkProvider.getOptionalState(level, random, pos);
         if (blockBelowTrunk != null) {
             trunkSetter.accept(pos, blockBelowTrunk);
         }
     }
 
-    protected boolean placeLog(WorldGenLevel level, BiConsumer<BlockPos, BlockState> trunkSetter, RandomSource random, BlockPos pos, TreeFeature tree) {
-        return this.placeLog(level, trunkSetter, random, pos, tree, Function.identity());
+    protected boolean placeLog(WorldGenLevel level, BiConsumer<BlockPos, BlockState> trunkSetter, RandomSource random, BlockPos pos, TreeConfiguration config) {
+        return this.placeLog(level, trunkSetter, random, pos, config, Function.identity());
     }
 
     protected boolean placeLog(
@@ -76,11 +_,11 @@
         BiConsumer<BlockPos, BlockState> trunkSetter,
         RandomSource random,
         BlockPos pos,
-        TreeFeature tree,
+        TreeConfiguration config,
         Function<BlockState, BlockState> stateModifier
     ) {
         if (this.validTreePos(level, pos)) {
-            trunkSetter.accept(pos, stateModifier.apply(tree.trunkProvider().getState(level, random, pos)));
+            trunkSetter.accept(pos, stateModifier.apply(config.trunkProvider.getState(level, random, pos)));
             return true;
         } else {
             return false;
@@ -88,10 +_,10 @@
     }
 
     protected void placeLogIfFree(
-        WorldGenLevel level, BiConsumer<BlockPos, BlockState> trunkSetter, RandomSource random, BlockPos.MutableBlockPos pos, TreeFeature tree
+        WorldGenLevel level, BiConsumer<BlockPos, BlockState> trunkSetter, RandomSource random, BlockPos.MutableBlockPos pos, TreeConfiguration config
     ) {
         if (this.isFree(level, pos)) {
-            this.placeLog(level, trunkSetter, random, pos, tree);
+            this.placeLog(level, trunkSetter, random, pos, config);
         }
     }
 
