--- a/net/minecraft/client/gui/screens/recipebook/GhostSlots.java
+++ b/net/minecraft/client/gui/screens/recipebook/GhostSlots.java
@@ -35,11 +_,11 @@
         }
     }
 
-    public void setInput(Slot slot, ContextMap context, SlotDisplay contents) {
+    protected void setInput(Slot slot, ContextMap context, SlotDisplay contents) {
         this.setSlot(slot, context, contents, false);
     }
 
-    public void setResult(Slot slot, ContextMap context, SlotDisplay contents) {
+    protected void setResult(Slot slot, ContextMap context, SlotDisplay contents) {
         this.setSlot(slot, context, contents, true);
     }
 
