--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -93,7 +_,14 @@
                     BlockState blockState = level.getBlockState(testPos);
                     if (blockState.isAir()) {
                         if (this.hasFlammableNeighbours(level, testPos)) {
-                            level.setBlockAndUpdate(testPos, BaseFireBlock.getState(level, testPos));
+                            // CraftBukkit start - Prevent lava putting something on fire
+                            if (level.getBlockState(testPos).getBlock() != Blocks.FIRE) {
+                                if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, testPos, pos).isCancelled()) {
+                                    continue;
+                                }
+                            }
+                            // CraftBukkit end
+                            level.setBlockAndUpdate(testPos, net.neoforged.neoforge.event.EventHooks.fireFluidPlaceBlockEvent(level, testPos, pos, BaseFireBlock.getState(level, testPos)));
                             return;
                         }
                     } else if (blockState.blocksMotion()) {
@@ -107,8 +_,16 @@
                         return;
                     }
 
-                    if (level.isEmptyBlock(testPos.above()) && this.isFlammable(level, testPos)) {
-                        level.setBlockAndUpdate(testPos.above(), BaseFireBlock.getState(level, testPos));
+                    if (level.isEmptyBlock(testPos.above()) && this.isFlammable(level, testPos, Direction.UP)) {
+                        // CraftBukkit start - Prevent lava putting something on fire
+                        BlockPos up = testPos.above();
+                        if (level.getBlockState(up).getBlock() != Blocks.FIRE) {
+                            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, up, pos).isCancelled()) {
+                                continue;
+                            }
+                        }
+                        // CraftBukkit end
+                        level.setBlockAndUpdate(testPos.above(), net.neoforged.neoforge.event.EventHooks.fireFluidPlaceBlockEvent(level, testPos.above(), pos, BaseFireBlock.getState(level, testPos)));
                     }
                 }
             }
@@ -117,14 +_,15 @@
 
     @Override
     protected void entityInside(Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
+        BlockPos savedPos = pos.immutable(); // Paper - track lava contact
         effectApplier.apply(InsideBlockEffectType.CLEAR_FREEZE);
         effectApplier.apply(InsideBlockEffectType.LAVA_IGNITE);
-        effectApplier.runAfter(InsideBlockEffectType.LAVA_IGNITE, Entity::lavaHurt);
+        effectApplier.runAfter(InsideBlockEffectType.LAVA_IGNITE, ignitedEntity -> ignitedEntity.lavaHurt(savedPos)); // Paper - track lava contact
     }
 
     private boolean hasFlammableNeighbours(LevelReader level, BlockPos pos) {
         for (Direction direction : Direction.values()) {
-            if (this.isFlammable(level, pos.relative(direction))) {
+            if (this.isFlammable(level, pos.relative(direction), direction.getOpposite())) {
                 return true;
             }
         }
@@ -132,10 +_,20 @@
         return false;
     }
 
+    /** @deprecated Forge: use {@link LavaFluid#isFlammable(LevelReader,BlockPos,Direction)} instead */
+    @Deprecated
     private boolean isFlammable(LevelReader level, BlockPos pos) {
         return level.isInsideBuildHeight(pos.getY()) && !level.hasChunkAt(pos) ? false : level.getBlockState(pos).ignitedByLava();
     }
 
+    private boolean isFlammable(LevelReader level, BlockPos pos, Direction face) {
+        if (level.isInsideBuildHeight(pos.getY()) && !level.hasChunkAt(pos)) {
+            return false;
+        }
+        BlockState state = level.getBlockState(pos);
+        return state.ignitedByLava(level, pos, face);
+    }
+
     @Override
     public @Nullable ParticleOptions getDripParticle() {
         return ParticleTypes.DRIPPING_LAVA;
@@ -206,7 +_,11 @@
             FluidState fluidState = level.getFluidState(pos);
             if (this.is(FluidTags.LAVA) && fluidState.is(FluidTags.WATER)) {
                 if (state.getBlock() instanceof LiquidBlock) {
-                    level.setBlock(pos, Blocks.STONE.defaultBlockState(), 3);
+                    // CraftBukkit start
+                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level.getMinecraftWorld(), pos, net.neoforged.neoforge.event.EventHooks.fireFluidPlaceBlockEvent(level, pos, pos, Blocks.STONE.defaultBlockState()), 3)) {
+                        return;
+                    }
+                    // CraftBukkit end
                 }
 
                 this.fizz(level, pos);
@@ -224,7 +_,7 @@
 
     @Override
     protected float getExplosionResistance() {
-        return 100.0F;
+        return Blocks.LAVA.getExplosionResistance(); // Paper - Get explosion resistance from actual block
     }
 
     @Override
