--- a/net/minecraft/world/level/material/MapColor.java
+++ b/net/minecraft/world/level/material/MapColor.java
@@ -3,7 +_,7 @@
 import com.google.common.base.Preconditions;
 
 public class MapColor {
-    public static final MapColor[] MATERIAL_COLORS = new MapColor[64];
+    private static final MapColor[] MATERIAL_COLORS = new MapColor[64];
     public static final MapColor NONE = new MapColor(0, 0);
     public static final MapColor GRASS = new MapColor(1, 8368696);
     public static final MapColor SAND = new MapColor(2, 16247203);
