--- a/net/minecraft/world/level/material/MapColor.java
+++ b/net/minecraft/world/level/material/MapColor.java
@@ -5,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 public class MapColor {
-    private static final @Nullable MapColor[] MATERIAL_COLORS = new MapColor[64];
+    public static final @Nullable MapColor[] MATERIAL_COLORS = new MapColor[64];
     public static final MapColor NONE = new MapColor(0, 0);
     public static final MapColor GRASS = new MapColor(1, 8368696);
     public static final MapColor SAND = new MapColor(2, 16247203);
