--- a/net/minecraft/util/datafix/DataFixers.java
+++ b/net/minecraft/util/datafix/DataFixers.java
@@ -46,7 +_,6 @@
 import net.minecraft.util.datafix.fixes.BiomeFix;
 import net.minecraft.util.datafix.fixes.BitStorageAlignFix;
 import net.minecraft.util.datafix.fixes.BlendingDataFix;
-import net.minecraft.util.datafix.fixes.BlendingDataNoValueHeightFix;
 import net.minecraft.util.datafix.fixes.BlendingDataRemoveFromNetherEndFix;
 import net.minecraft.util.datafix.fixes.BlockEntityBannerColorFix;
 import net.minecraft.util.datafix.fixes.BlockEntityBlockStateFix;
@@ -61,7 +_,8 @@
 import net.minecraft.util.datafix.fixes.BlockEntityUUIDFix;
 import net.minecraft.util.datafix.fixes.BlockNameFlatteningFix;
 import net.minecraft.util.datafix.fixes.BlockPosFormatAndRenamesFix;
-import net.minecraft.util.datafix.fixes.BlockStateFieldNamesFix;
+import net.minecraft.util.datafix.fixes.BlockPropertyRenameAndFix;
+import net.minecraft.util.datafix.fixes.BlockRenameFix;
 import net.minecraft.util.datafix.fixes.BlockStateStructureTemplateFix;
 import net.minecraft.util.datafix.fixes.BoatSplitFix;
 import net.minecraft.util.datafix.fixes.CatTypeFix;
@@ -138,7 +_,6 @@
 import net.minecraft.util.datafix.fixes.EntityZombifiedPiglinRenameFix;
 import net.minecraft.util.datafix.fixes.EquipmentFormatFix;
 import net.minecraft.util.datafix.fixes.EquippableAssetRenameFix;
-import net.minecraft.util.datafix.fixes.ExplorerMapItemFix;
 import net.minecraft.util.datafix.fixes.FeatureFlagRemoveFix;
 import net.minecraft.util.datafix.fixes.FilteredBooksFix;
 import net.minecraft.util.datafix.fixes.FilteredSignsFix;
@@ -175,14 +_,12 @@
 import net.minecraft.util.datafix.fixes.ItemStackUUIDFix;
 import net.minecraft.util.datafix.fixes.ItemWaterPotionFix;
 import net.minecraft.util.datafix.fixes.JigsawPropertiesFix;
+import net.minecraft.util.datafix.fixes.JigsawRotationFix;
 import net.minecraft.util.datafix.fixes.JukeboxTicksSinceSongStartedFix;
 import net.minecraft.util.datafix.fixes.LeavesFix;
-import net.minecraft.util.datafix.fixes.LegacyBlockPropertyRenameAndFix;
-import net.minecraft.util.datafix.fixes.LegacyBlockRenameFix;
 import net.minecraft.util.datafix.fixes.LegacyDimensionIdFix;
 import net.minecraft.util.datafix.fixes.LegacyDragonFightFix;
 import net.minecraft.util.datafix.fixes.LegacyHoverEventFix;
-import net.minecraft.util.datafix.fixes.LegacyJigsawRotationFix;
 import net.minecraft.util.datafix.fixes.LegacyWorldBorderFix;
 import net.minecraft.util.datafix.fixes.LevelDatDifficultyFix;
 import net.minecraft.util.datafix.fixes.LevelDatToSavedDataPreparationFix;
@@ -193,10 +_,8 @@
 import net.minecraft.util.datafix.fixes.LockComponentPredicateFix;
 import net.minecraft.util.datafix.fixes.LodestoneCompassComponentFix;
 import net.minecraft.util.datafix.fixes.MapBannerBlockPosFormatFix;
-import net.minecraft.util.datafix.fixes.MapColorComponentRemovalFix;
 import net.minecraft.util.datafix.fixes.MapIdFix;
 import net.minecraft.util.datafix.fixes.MemoryExpiryDataFix;
-import net.minecraft.util.datafix.fixes.MergeTerrainChunkStatusFix;
 import net.minecraft.util.datafix.fixes.MissingDimensionFix;
 import net.minecraft.util.datafix.fixes.MobEffectIdFix;
 import net.minecraft.util.datafix.fixes.MobSpawnerEntityIdentifiersFix;
@@ -243,7 +_,6 @@
 import net.minecraft.util.datafix.fixes.References;
 import net.minecraft.util.datafix.fixes.RemapChunkStatusFix;
 import net.minecraft.util.datafix.fixes.RemoveBlockEntityTagFix;
-import net.minecraft.util.datafix.fixes.RemoveBlockTransformerComponentFix;
 import net.minecraft.util.datafix.fixes.RemoveEmptyItemInBrushableBlockFix;
 import net.minecraft.util.datafix.fixes.RemoveGolemGossipFix;
 import net.minecraft.util.datafix.fixes.RenameEnchantmentsFix;
@@ -257,7 +_,6 @@
 import net.minecraft.util.datafix.fixes.ScoreboardDisplayNameFix;
 import net.minecraft.util.datafix.fixes.ScoreboardDisplaySlotFix;
 import net.minecraft.util.datafix.fixes.SignTextStrictJsonFix;
-import net.minecraft.util.datafix.fixes.SignsOpFlagFix;
 import net.minecraft.util.datafix.fixes.SpawnerDataFix;
 import net.minecraft.util.datafix.fixes.StatsCounterFix;
 import net.minecraft.util.datafix.fixes.StatsRenameFix;
@@ -266,7 +_,6 @@
 import net.minecraft.util.datafix.fixes.StructureReferenceCountFix;
 import net.minecraft.util.datafix.fixes.StructureSettingsFlattenFix;
 import net.minecraft.util.datafix.fixes.StructuresBecomeConfiguredFix;
-import net.minecraft.util.datafix.fixes.SwingAnimationComponentSplitFix;
 import net.minecraft.util.datafix.fixes.TextComponentHoverAndClickEventFix;
 import net.minecraft.util.datafix.fixes.TextComponentStringifiedFlagsFix;
 import net.minecraft.util.datafix.fixes.ThrownPotionSplitFix;
@@ -276,7 +_,6 @@
 import net.minecraft.util.datafix.fixes.TrialSpawnerConfigFix;
 import net.minecraft.util.datafix.fixes.TrialSpawnerConfigInRegistryFix;
 import net.minecraft.util.datafix.fixes.TridentAnimationFix;
-import net.minecraft.util.datafix.fixes.UnfilledBuriedTreasureMapFix;
 import net.minecraft.util.datafix.fixes.UnflattenTextComponentFix;
 import net.minecraft.util.datafix.fixes.VariantRenameFix;
 import net.minecraft.util.datafix.fixes.VillagerDataFix;
@@ -408,7 +_,6 @@
 import net.minecraft.util.datafix.schemas.V4996;
 import net.minecraft.util.datafix.schemas.V4996_1;
 import net.minecraft.util.datafix.schemas.V4997;
-import net.minecraft.util.datafix.schemas.V5000;
 import net.minecraft.util.datafix.schemas.V501;
 import net.minecraft.util.datafix.schemas.V700;
 import net.minecraft.util.datafix.schemas.V701;
@@ -568,6 +_,20 @@
         Schema v1456 = fixerUpper.addSchema(1456, SAME_NAMESPACED);
         fixerUpper.addFixer(new EntityItemFrameDirectionFix(v1456, false));
         Schema v1458 = fixerUpper.addSchema(1458, V1458::new);
+        // CraftBukkit start
+        fixerUpper.addFixer(new com.mojang.datafixers.DataFix(v1458, false) {
+            @Override
+            protected com.mojang.datafixers.TypeRewriteRule makeRule() {
+                return this.fixTypeEverywhereTyped("Player CustomName", this.getInputSchema().getType(References.PLAYER), (typed) -> {
+                    return typed.update(DSL.remainderFinder(), (dynamic) -> {
+                        String s = dynamic.get("CustomName").asString("");
+
+                        return s.isEmpty() ? dynamic.remove("CustomName") : dynamic.set("CustomName", LegacyComponentDataFixUtils.createPlainTextComponent(dynamic.getOps(), s));
+                    });
+                });
+            }
+        });
+        // CraftBukkit end
         fixerUpper.addFixer(new EntityCustomNameToComponentFix(v1458));
         fixerUpper.addFixer(new ItemCustomNameToComponentFix(v1458));
         fixerUpper.addFixer(new BlockEntityCustomNameToComponentFix(v1458));
@@ -581,7 +_,7 @@
         Schema v1474 = fixerUpper.addSchema(1474, SAME_NAMESPACED);
         fixerUpper.addFixer(new ColorlessShulkerEntityFix(v1474, false));
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v1474,
                 "Colorless shulker block fixer",
                 block -> Objects.equals(NamespacedSchema.ensureNamespaced(block), "minecraft:purple_shulker_box") ? "minecraft:shulker_box" : block
@@ -596,14 +_,14 @@
         );
         Schema v1475 = fixerUpper.addSchema(1475, SAME_NAMESPACED);
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v1475,
                 "Flowing fixer",
                 createRenamer(ImmutableMap.of("minecraft:flowing_water", "minecraft:water", "minecraft:flowing_lava", "minecraft:lava"))
             )
         );
         Schema v1480 = fixerUpper.addSchema(1480, SAME_NAMESPACED);
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v1480, "Rename coral blocks", createRenamer(RenamedCoralFix.RENAMED_IDS)));
+        fixerUpper.addFixer(BlockRenameFix.create(v1480, "Rename coral blocks", createRenamer(RenamedCoralFix.RENAMED_IDS)));
         fixerUpper.addFixer(ItemRenameFix.create(v1480, "Rename coral items", createRenamer(RenamedCoralFix.RENAMED_IDS)));
         Schema v1481 = fixerUpper.addSchema(1481, V1481::new);
         fixerUpper.addFixer(new AddNewChoices(v1481, "Add conduit", References.BLOCK_ENTITY));
@@ -619,7 +_,7 @@
             )
         );
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v1484,
                 "Rename seagrass blocks",
                 createRenamer(ImmutableMap.of("minecraft:sea_grass", "minecraft:seagrass", "minecraft:tall_sea_grass", "minecraft:tall_seagrass"))
@@ -645,7 +_,7 @@
             )
         );
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v1487,
                 "Rename prismarine_brick(s)_* items",
                 createRenamer(
@@ -660,7 +_,7 @@
         );
         Schema v1488 = fixerUpper.addSchema(1488, V1488::new);
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v1488, "Rename kelp/kelptop", createRenamer(ImmutableMap.of("minecraft:kelp_top", "minecraft:kelp", "minecraft:kelp", "minecraft:kelp_plant"))
             )
         );
@@ -703,7 +_,7 @@
         );
         fixerUpper.addFixer(new IglooMetadataRemovalFix(v1488, false));
         Schema v1490 = fixerUpper.addSchema(1490, SAME_NAMESPACED);
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v1490, "Rename melon_block", createRenamer("minecraft:melon_block", "minecraft:melon")));
+        fixerUpper.addFixer(BlockRenameFix.create(v1490, "Rename melon_block", createRenamer("minecraft:melon_block", "minecraft:melon")));
         fixerUpper.addFixer(
             ItemRenameFix.create(
                 v1490,
@@ -737,7 +_,7 @@
         Schema v1506 = fixerUpper.addSchema(1506, SAME_NAMESPACED);
         fixerUpper.addFixer(new LevelDataGeneratorOptionsFix(v1506, false));
         Schema v1510 = fixerUpper.addSchema(1510, V1510::new);
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v1510, "Block renamening fix", createRenamer(EntityTheRenameningFix.RENAMED_BLOCKS)));
+        fixerUpper.addFixer(BlockRenameFix.create(v1510, "Block renamening fix", createRenamer(EntityTheRenameningFix.RENAMED_BLOCKS)));
         fixerUpper.addFixer(ItemRenameFix.create(v1510, "Item renamening fix", createRenamer(EntityTheRenameningFix.RENAMED_ITEMS)));
         fixerUpper.addFixer(new NamespacedTypeRenameFix(v1510, "Recipes renamening fix", References.RECIPE, createRenamer(RecipesRenameningFix.RECIPES)));
         fixerUpper.addFixer(new EntityTheRenameningFix(v1510, true));
@@ -753,7 +_,7 @@
         fixerUpper.addFixer(new ScoreboardDisplayNameFix(v1514, "TeamDisplayNameFix", References.TEAM));
         fixerUpper.addFixer(new ObjectiveRenderTypeFix(v1514));
         Schema v1515 = fixerUpper.addSchema(1515, SAME_NAMESPACED);
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v1515, "Rename coral fan blocks", createRenamer(RenamedCoralFansFix.RENAMED_IDS)));
+        fixerUpper.addFixer(BlockRenameFix.create(v1515, "Rename coral fan blocks", createRenamer(RenamedCoralFansFix.RENAMED_IDS)));
         Schema v1624 = fixerUpper.addSchema(1624, SAME_NAMESPACED);
         fixerUpper.addFixer(new TrappedChestBlockEntityFix(v1624, false));
         Schema v1800 = fixerUpper.addSchema(1800, V1800::new);
@@ -763,7 +_,7 @@
         fixerUpper.addFixer(new AddNewChoices(v1801, "Added Illager Beast", References.ENTITY));
         Schema v1802 = fixerUpper.addSchema(1802, SAME_NAMESPACED);
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v1802,
                 "Rename sign blocks & stone slabs",
                 createRenamer(
@@ -849,7 +_,7 @@
         UnaryOperator<String> beehiveRenamer = createRenamer("minecraft:bee_hive", "minecraft:beehive");
         fixerUpper.addFixer(ItemRenameFix.create(v2209, "Rename bee_hive item to beehive", beehiveRenamer));
         fixerUpper.addFixer(new PoiTypeRenameFix(v2209, "Rename bee_hive poi to beehive", beehiveRenamer));
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v2209, "Rename bee_hive block to beehive", beehiveRenamer));
+        fixerUpper.addFixer(BlockRenameFix.create(v2209, "Rename bee_hive block to beehive", beehiveRenamer));
         Schema v2211 = fixerUpper.addSchema(2211, SAME_NAMESPACED);
         fixerUpper.addFixer(new StructureReferenceCountFix(v2211, false));
         Schema v2218 = fixerUpper.addSchema(2218, SAME_NAMESPACED);
@@ -877,7 +_,7 @@
             )
         );
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v2508,
                 "Renamed fungi blocks to fungus",
                 createRenamer(ImmutableMap.of("minecraft:warped_fungi", "minecraft:warped_fungus", "minecraft:crimson_fungi", "minecraft:crimson_fungus"))
@@ -902,7 +_,7 @@
         fixerUpper.addFixer(new GossipUUIDFix(v2516, "minecraft:zombie_villager"));
         Schema v2518 = fixerUpper.addSchema(2518, SAME_NAMESPACED);
         fixerUpper.addFixer(new JigsawPropertiesFix(v2518, false));
-        fixerUpper.addFixer(new LegacyJigsawRotationFix(v2518));
+        fixerUpper.addFixer(new JigsawRotationFix(v2518));
         Schema v2519 = fixerUpper.addSchema(2519, V2519::new);
         fixerUpper.addFixer(new AddNewChoices(v2519, "Added Strider", References.ENTITY));
         Schema v2522 = fixerUpper.addSchema(2522, V2522::new);
@@ -947,7 +_,7 @@
             )
         );
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v2528,
                 "Rename soul fire torch and soul fire lantern",
                 createRenamer(
@@ -997,9 +_,7 @@
         fixerUpper.addFixer(new CauldronRenameFix(v2679, false));
         Schema v2680 = fixerUpper.addSchema(2680, SAME_NAMESPACED);
         fixerUpper.addFixer(ItemRenameFix.create(v2680, "Renamed grass path item to dirt path", createRenamer("minecraft:grass_path", "minecraft:dirt_path")));
-        fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(v2680, "Renamed grass path block to dirt path", createRenamer("minecraft:grass_path", "minecraft:dirt_path"))
-        );
+        fixerUpper.addFixer(BlockRenameFix.create(v2680, "Renamed grass path block to dirt path", createRenamer("minecraft:grass_path", "minecraft:dirt_path")));
         Schema v2684 = fixerUpper.addSchema(2684, V2684::new);
         fixerUpper.addFixer(new AddNewChoices(v2684, "Added Sculk Sensor", References.BLOCK_ENTITY));
         Schema v2686 = fixerUpper.addSchema(2686, V2686::new);
@@ -1031,7 +_,7 @@
             .put("minecraft:waxed_lightly_weathered_cut_copper_slab", "minecraft:waxed_exposed_cut_copper_slab")
             .build();
         fixerUpper.addFixer(ItemRenameFix.create(v2690, "Renamed copper block items to new oxidized terms", createRenamer(copperBlocksToRename)));
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v2690, "Renamed copper blocks to new oxidized terms", createRenamer(copperBlocksToRename)));
+        fixerUpper.addFixer(BlockRenameFix.create(v2690, "Renamed copper blocks to new oxidized terms", createRenamer(copperBlocksToRename)));
         Schema v2691 = fixerUpper.addSchema(2691, SAME_NAMESPACED);
         ImmutableMap<String, String> waxedCopperRename = ImmutableMap.<String, String>builder()
             .put("minecraft:waxed_copper", "minecraft:waxed_copper_block")
@@ -1040,7 +_,7 @@
             .put("minecraft:exposed_copper_block", "minecraft:exposed_copper")
             .build();
         fixerUpper.addFixer(ItemRenameFix.create(v2691, "Rename copper item suffixes", createRenamer(waxedCopperRename)));
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v2691, "Rename copper blocks suffixes", createRenamer(waxedCopperRename)));
+        fixerUpper.addFixer(BlockRenameFix.create(v2691, "Rename copper blocks suffixes", createRenamer(waxedCopperRename)));
         Schema v2693 = fixerUpper.addSchema(2693, SAME_NAMESPACED);
         fixerUpper.addFixer(new AddFlagIfNotPresentFix(v2693, References.WORLD_GEN_SETTINGS, "has_increased_height_already", false));
         Schema v2696 = fixerUpper.addSchema(2696, SAME_NAMESPACED);
@@ -1064,10 +_,10 @@
             .put("minecraft:chiseled_grimstone", "minecraft:chiseled_deepslate")
             .build();
         fixerUpper.addFixer(ItemRenameFix.create(v2696, "Renamed grimstone block items to deepslate", createRenamer(grimstoneBlocksToRename)));
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v2696, "Renamed grimstone blocks to deepslate", createRenamer(grimstoneBlocksToRename)));
+        fixerUpper.addFixer(BlockRenameFix.create(v2696, "Renamed grimstone blocks to deepslate", createRenamer(grimstoneBlocksToRename)));
         Schema v2700 = fixerUpper.addSchema(2700, SAME_NAMESPACED);
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v2700,
                 "Renamed cave vines blocks",
                 createRenamer(ImmutableMap.of("minecraft:cave_vines_head", "minecraft:cave_vines", "minecraft:cave_vines_body", "minecraft:cave_vines_plant"))
@@ -1093,7 +_,7 @@
             )
         );
         fixerUpper.addFixer(
-            LegacyBlockRenameFix.create(
+            BlockRenameFix.create(
                 v2717,
                 "Rename azalea_leaves_flowers items",
                 createRenamer(ImmutableMap.of("minecraft:azalea_leaves_flowers", "minecraft:flowering_azalea_leaves"))
@@ -1386,17 +_,42 @@
         fixerUpper.addFixer(new AddNewChoices(v3689, "Added Trial Spawner", References.BLOCK_ENTITY));
         Schema v3692 = fixerUpper.addSchema(3692, SAME_NAMESPACED);
         UnaryOperator<String> grassRenamer = createRenamer(Map.of("minecraft:grass", "minecraft:short_grass"));
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v3692, "Rename grass block to short_grass", grassRenamer));
+        fixerUpper.addFixer(BlockRenameFix.create(v3692, "Rename grass block to short_grass", grassRenamer));
         fixerUpper.addFixer(ItemRenameFix.create(v3692, "Rename grass item to short_grass", grassRenamer));
         Schema v3799 = fixerUpper.addSchema(3799, V3799::new);
         fixerUpper.addFixer(new AddNewChoices(v3799, "Added Armadillo", References.ENTITY));
         Schema v3800 = fixerUpper.addSchema(3800, SAME_NAMESPACED);
         UnaryOperator<String> scuteRenamer = createRenamer(Map.of("minecraft:scute", "minecraft:turtle_scute"));
         fixerUpper.addFixer(ItemRenameFix.create(v3800, "Rename scute item to turtle_scute", scuteRenamer));
+
+        // Neo: Remove step height attribute, which cannot be sanely updated
+        // Happens in 23w51a
+        Schema neoSchema3801 = fixerUpper.addSchema(3801, SAME_NAMESPACED);
+        fixerUpper.addFixer(new net.neoforged.neoforge.common.data.fixes.NeoForgeEntityLegacyAttributesFix(
+                "(Neo) Remove step height attribute",
+                neoSchema3801,
+                java.util.List.of(
+                        "neoforge:step_height"
+                )
+        ));
+
         Schema v3803 = fixerUpper.addSchema(3803, SAME_NAMESPACED);
         fixerUpper.addFixer(
             new RenameEnchantmentsFix(v3803, "Rename sweeping enchant to sweeping_edge", Map.of("minecraft:sweeping", "minecraft:sweeping_edge"))
         );
+
+        // Neo: rename neo attributes to new MC attributes
+        // Happens in 24w03a
+        Schema neoSchema3804 = fixerUpper.addSchema(3804, SAME_NAMESPACED);
+        fixerUpper.addFixer(new AttributesRenameLegacy(
+                neoSchema3804,
+                "(Neo) Rename reach attributes to vanilla",
+                createRenamer(ImmutableMap.of(
+                        "neoforge:entity_reach", "minecraft:player.entity_interaction_range",
+                        "neoforge:block_reach", "minecraft:player.block_interaction_range"
+                ))
+        ));
+
         Schema v3807 = fixerUpper.addSchema(3807, V3807::new);
         fixerUpper.addFixer(new AddNewChoices(v3807, "Added Vault", References.BLOCK_ENTITY));
         Schema v3807_1 = fixerUpper.addSchema(3807, 1, SAME_NAMESPACED);
@@ -1419,6 +_,18 @@
                 v3814, "Rename jump strength attribute", createRenamer("minecraft:horse.jump_strength", "minecraft:generic.jump_strength")
             )
         );
+
+        // Neo: rename neo attributes to new MC attributes
+        // Happens in 24w06a
+        Schema neoSchema3815 = fixerUpper.addSchema(3815, SAME_NAMESPACED);
+        fixerUpper.addFixer(new AttributesRenameLegacy(
+                neoSchema3815,
+                "(Neo) Rename gravity attribute to vanilla",
+                createRenamer(
+                        "neoforge:entity_gravity", "minecraft:generic.gravity"
+                )
+        ));
+
         Schema v3816 = fixerUpper.addSchema(3816, V3816::new);
         fixerUpper.addFixer(new AddNewChoices(v3816, "Added Bogged", References.ENTITY));
         Schema v3818 = fixerUpper.addSchema(3818, V3818::new);
@@ -1530,7 +_,7 @@
         fixerUpper.addFixer(new DropChancesFormatFix(v4293));
         Schema v4294 = fixerUpper.addSchema(4294, SAME_NAMESPACED);
         fixerUpper.addFixer(
-            new LegacyBlockPropertyRenameAndFix(
+            new BlockPropertyRenameAndFix(
                 v4294,
                 "CreakingHeartBlockStateFix",
                 "minecraft:creaking_heart",
@@ -1555,7 +_,7 @@
         fixerUpper.addFixer(new EntityFallDistanceFloatToDoubleFix(v4303, References.ENTITY));
         fixerUpper.addFixer(new EntityFallDistanceFloatToDoubleFix(v4303, References.PLAYER));
         Schema v4305 = fixerUpper.addSchema(4305, SAME_NAMESPACED);
-        fixerUpper.addFixer(new LegacyBlockPropertyRenameAndFix(v4305, "rename test block mode", "minecraft:test_block", "test_block_mode", "mode", a -> a));
+        fixerUpper.addFixer(new BlockPropertyRenameAndFix(v4305, "rename test block mode", "minecraft:test_block", "test_block_mode", "mode", a -> a));
         Schema v4306 = fixerUpper.addSchema(4306, V4306::new);
         fixerUpper.addFixer(new ThrownPotionSplitFix(v4306));
         Schema v4307 = fixerUpper.addSchema(4307, V4307::new);
@@ -1592,7 +_,7 @@
         Schema v4537 = fixerUpper.addSchema(4537, SAME_NAMESPACED);
         fixerUpper.addFixer(new ChunkDeleteLightFix(v4537));
         Schema v4541 = fixerUpper.addSchema(4541, SAME_NAMESPACED);
-        fixerUpper.addFixer(LegacyBlockRenameFix.create(v4541, "Rename chain to iron_chain", createRenamer("minecraft:chain", "minecraft:iron_chain")));
+        fixerUpper.addFixer(BlockRenameFix.create(v4541, "Rename chain to iron_chain", createRenamer("minecraft:chain", "minecraft:iron_chain")));
         fixerUpper.addFixer(ItemRenameFix.create(v4541, "Rename chain to iron_chain", createRenamer("minecraft:chain", "minecraft:iron_chain")));
         Schema v4543 = fixerUpper.addSchema(4543, V4543::new);
         fixerUpper.addFixer(new AddNewChoices(v4543, "Added Mannequin", References.ENTITY));
@@ -1664,46 +_,6 @@
         fixerUpper.addFixer(new BlendingDataFix(blendingSchema));
         Schema v4997 = fixerUpper.addSchema(4997, V4997::new);
         fixerUpper.addFixer(new AddNewChoices(v4997, "Added Poplar Boat and Poplar Chest Boat", References.ENTITY));
-        Schema v5000 = fixerUpper.addSchema(5000, V5000::new);
-        fixerUpper.addFixer(new AddNewChoices(v5000, "Added Cushion", References.ENTITY));
-        Schema v5002 = fixerUpper.addSchema(5002, SAME_NAMESPACED);
-        fixerUpper.addFixer(new SignsOpFlagFix(v5002));
-        Schema v5006 = fixerUpper.addSchema(5006, SAME_NAMESPACED);
-        fixerUpper.addFixer(new BlockStateFieldNamesFix(v5006));
-        Schema v5007 = fixerUpper.addSchema(5007, SAME_NAMESPACED);
-        fixerUpper.addFixer(new SwingAnimationComponentSplitFix(v5007));
-        Schema v5008 = fixerUpper.addSchema(5008, SAME_NAMESPACED);
-        fixerUpper.addFixer(new ExplorerMapItemFix(v5008));
-        fixerUpper.addFixer(new MapColorComponentRemovalFix(v5008));
-        Schema v5009 = fixerUpper.addSchema(5009, SAME_NAMESPACED);
-        fixerUpper.addFixer(new BlendingDataNoValueHeightFix(v5009));
-        Schema v5012 = fixerUpper.addSchema(5012, SAME_NAMESPACED);
-        fixerUpper.addFixer(
-            ItemRenameFix.create(
-                v5012,
-                "Rename explorer map items",
-                createRenamer(
-                    ImmutableMap.of(
-                        "minecraft:ocean_explorer_map",
-                        "minecraft:ocean_monument_map",
-                        "minecraft:swamp_explorer_map",
-                        "minecraft:swamp_hut_map",
-                        "minecraft:trial_explorer_map",
-                        "minecraft:buried_trial_chambers_map",
-                        "minecraft:woodland_explorer_map",
-                        "minecraft:woodland_mansion_map",
-                        "minecraft:jungle_explorer_map",
-                        "minecraft:jungle_pyramid_map"
-                    )
-                )
-            )
-        );
-        Schema v5013 = fixerUpper.addSchema(5013, SAME_NAMESPACED);
-        fixerUpper.addFixer(new MergeTerrainChunkStatusFix(v5013));
-        Schema v5014 = fixerUpper.addSchema(5014, SAME_NAMESPACED);
-        fixerUpper.addFixer(new RemoveBlockTransformerComponentFix(v5014));
-        Schema v5016 = fixerUpper.addSchema(5016, SAME_NAMESPACED);
-        fixerUpper.addFixer(new UnfilledBuriedTreasureMapFix(v5016));
     }
 
     private static UnaryOperator<String> createRenamerNoNamespace(Map<String, String> map) {
