--- a/net/minecraft/util/datafix/ExtraDataFixUtils.java
+++ b/net/minecraft/util/datafix/ExtraDataFixUtils.java
@@ -11,10 +_,15 @@
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.DynamicOps;
 import java.util.BitSet;
+import java.util.Map;
 import java.util.Optional;
 import java.util.function.Function;
 import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
 import java.util.stream.IntStream;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.NbtOps;
+import net.minecraft.nbt.Tag;
 import net.minecraft.util.Util;
 
 public class ExtraDataFixUtils {
@@ -71,12 +_,29 @@
         };
     }
 
+    public static Dynamic<?> blockState(String id, Map<String, String> properties) {
+        Dynamic<Tag> dynamic = new Dynamic<>(NbtOps.INSTANCE, new CompoundTag());
+        Dynamic<Tag> blockState = dynamic.set("Name", dynamic.createString(id));
+        if (!properties.isEmpty()) {
+            blockState = blockState.set(
+                "Properties",
+                dynamic.createMap(
+                    properties.entrySet()
+                        .stream()
+                        .collect(Collectors.toMap(entry -> dynamic.createString(entry.getKey()), entry -> dynamic.createString(entry.getValue())))
+                )
+            );
+        }
+
+        return blockState;
+    }
+
+    public static Dynamic<?> blockState(String id) {
+        return blockState(id, Map.of());
+    }
+
     public static Dynamic<?> fixStringField(Dynamic<?> dynamic, String fieldName, UnaryOperator<String> fix) {
         return dynamic.update(fieldName, field -> DataFixUtils.orElse(field.asString().map(fix).map(dynamic::createString).result(), field));
-    }
-
-    public static Optional<String> getPlainTranslationKey(Dynamic<?> textComponent) {
-        return textComponent.getMapValues().result().filter(fields -> fields.size() == 1).flatMap(fields -> textComponent.get("translate").asString().result());
     }
 
     public static String dyeColorIdToName(int id) {
