--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -277,6 +_,13 @@
 
             for (ItemStack itemstack : list) {
                 ItemEntity itementity = new ItemEntity(serverlevel, (double)blockpos.getX(), (double)blockpos.getY(), (double)blockpos.getZ(), itemstack);
+                // CraftBukkit start - handle EntityDropItemEvent
+                org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(this.getBukkitEntity(), (org.bukkit.entity.Item) itementity.getBukkitEntity());
+                org.bukkit.Bukkit.getPluginManager().callEvent(event);
+                if (event.isCancelled()) {
+                    continue;
+                }
+                // CraftBukkit end
                 itementity.setDefaultPickUpDelay();
                 serverlevel.addFreshEntity(itementity);
             }
@@ -299,7 +_,7 @@
                 if (this.tickCount % 10 == 0) {
                     this.level()
                         .playLocalSound(
-                            this.getX(), this.getY(), this.getZ(), blockstate.getSoundType().getHitSound(), this.getSoundSource(), 0.5F, 0.5F, false
+                            this.getX(), this.getY(), this.getZ(), blockstate.getSoundType(level(), blockpos.below(), this).getHitSound(), this.getSoundSource(), 0.5F, 0.5F, false
                         );
                 }
             }
@@ -337,10 +_,14 @@
 
     @Override
     public void spawnChildFromBreeding(ServerLevel p_277923_, Animal p_277857_) {
+        // Paper start - Add EntityFertilizeEggEvent event
+        final io.papermc.paper.event.entity.EntityFertilizeEggEvent result = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityFertilizeEggEvent(this, p_277857_);
+        if (result.isCancelled()) return;
+        // Paper end - Add EntityFertilizeEggEvent event
         ItemStack itemstack = new ItemStack(Items.SNIFFER_EGG);
         ItemEntity itementity = new ItemEntity(p_277923_, this.position().x(), this.position().y(), this.position().z(), itemstack);
         itementity.setDefaultPickUpDelay();
-        this.finalizeSpawnChildFromBreeding(p_277923_, p_277857_, null);
+        this.finalizeSpawnChildFromBreeding(p_277923_, p_277857_, null, result.getExperience()); // Paper - Add EntityFertilizeEggEvent event
         this.playSound(SoundEvents.SNIFFER_EGG_PLOP, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 0.5F);
         p_277923_.addFreshEntity(itementity);
     }
@@ -403,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.SNIFFER_DEATH;
     }
 
