--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -297,7 +_,7 @@
                 if (this.tickCount % 10 == 0) {
                     this.level()
                         .playLocalSound(
-                            this.getX(), this.getY(), this.getZ(), blockstate.getSoundType().getHitSound(), this.getSoundSource(), 0.5F, 0.5F, false
+                            this.getX(), this.getY(), this.getZ(), blockstate.getSoundType(level(), blockpos.below(), this).getHitSound(), this.getSoundSource(), 0.5F, 0.5F, false
                         );
                 }
             }
@@ -391,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return Set.of(Sniffer.State.DIGGING, Sniffer.State.SEARCHING).contains(this.getState()) ? null : SoundEvents.SNIFFER_IDLE;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_273718_) {
+    public SoundEvent getHurtSound(DamageSource p_273718_) {
         return SoundEvents.SNIFFER_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.SNIFFER_DEATH;
     }
 
