--- a/net/minecraft/world/entity/animal/sniffer/SnifferAi.java
+++ b/net/minecraft/world/entity/animal/sniffer/SnifferAi.java
@@ -101,6 +_,7 @@
                     }
                 }),
                 Pair.of(1, new FollowTemptation(sniffer -> 1.25F, sniffer -> sniffer.isBaby() ? 2.5 : 3.5) {
+                    @Override
                     protected void start(ServerLevel level, PathfinderMob body, long timestamp) {
                         SnifferAi.resetSniffing((Sniffer)body);
                         super.start(level, body, timestamp);
@@ -126,7 +_,7 @@
         );
     }
 
-    public static void updateActivity(Sniffer body) {
+    static void updateActivity(Sniffer body) {
         body.getBrain().setActiveActivityToFirstValid(ImmutableList.of(Activity.DIG, Activity.SNIFF, Activity.IDLE));
     }
 
