--- a/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/coppergolem/CopperGolem.java
@@ -385,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_435524_) {
+    public SoundEvent getHurtSound(DamageSource p_435524_) {
         return CopperGolemOxidationLevels.getOxidationLevel(this.getWeatherState()).hurtSound();
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return CopperGolemOxidationLevels.getOxidationLevel(this.getWeatherState()).deathSound();
     }
 
