--- a/net/minecraft/world/entity/projectile/windcharge/AbstractWindCharge.java
+++ b/net/minecraft/world/entity/projectile/windcharge/AbstractWindCharge.java
@@ -72,7 +_,7 @@
     }
 
     @Override
-    protected boolean canHitEntity(Entity p_326159_) {
+    public boolean canHitEntity(Entity p_326159_) {
         if (p_326159_ instanceof AbstractWindCharge) {
             return false;
         } else {
@@ -100,8 +_,7 @@
     }
 
     @Override
-    public void push(double p_334071_, double p_333979_, double p_333996_) {
-    }
+    public void push(double p_334071_, double p_333979_, double p_333996_, @org.jetbrains.annotations.Nullable Entity pushingEntity) {} // Paper - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
 
     public abstract void explode(Vec3 p_352265_);
 
@@ -113,7 +_,7 @@
             Vec3 vec3 = Vec3.atLowerCornerOf(vec3i).multiply(0.25, 0.25, 0.25);
             Vec3 vec31 = p_325933_.getLocation().add(vec3);
             this.explode(vec31);
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
 
@@ -121,7 +_,7 @@
     protected void onHit(HitResult p_326337_) {
         super.onHit(p_326337_);
         if (!this.level().isClientSide) {
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
 
@@ -155,7 +_,7 @@
     public void tick() {
         if (!this.level().isClientSide && this.getBlockY() > this.level().getMaxBuildHeight() + 30) {
             this.explode(this.position());
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.OUT_OF_WORLD); // CraftBukkit - add Bukkit remove cause
         } else {
             super.tick();
         }
