--- a/net/minecraft/world/entity/projectile/windcharge/AbstractWindCharge.java
+++ b/net/minecraft/world/entity/projectile/windcharge/AbstractWindCharge.java
@@ -103,7 +_,7 @@
     public void push(double p_334071_, double p_333979_, double p_333996_) {
     }
 
-    public abstract void explode(Vec3 p_352265_);
+    protected abstract void explode(Vec3 p_352265_);
 
     @Override
     protected void onHitBlock(BlockHitResult p_325933_) {
