--- a/net/minecraft/world/entity/monster/zombie/Drowned.java
+++ b/net/minecraft/world/entity/monster/zombie/Drowned.java
@@ -1,10 +_,8 @@
 package net.minecraft.world.entity.monster.zombie;
 
-import com.google.common.annotations.VisibleForTesting;
 import java.util.EnumSet;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
-import net.minecraft.core.component.DataComponents;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
@@ -73,7 +_,6 @@
         .withEyeHeight(0.775F)
         .withAttachments(EntityAttachments.builder().attach(EntityAttachment.VEHICLE, 0.0F, 0.1875F, 0.0F));
     private boolean searchingForLand;
-    private float rangedAttackUncertainty = -1.0F;
 
     public Drowned(EntityType<? extends Drowned> type, Level level) {
         super(type, level);
@@ -81,6 +_,74 @@
         this.setPathfindingMalus(PathType.WATER, 0.0F);
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.drownedRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.drownedRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.drownedControllable;
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.drownedMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.drownedScale);
+    }
+
+    @Override
+    protected void randomizeReinforcementsChance() {
+        this.getAttribute(Attributes.SPAWN_REINFORCEMENTS_CHANCE).setBaseValue(this.random.nextDouble() * this.level().purpurConfig.drownedSpawnReinforcements);
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Configurable jockey options
+    @Override
+    public boolean jockeyOnlyBaby() {
+        return level().purpurConfig.drownedJockeyOnlyBaby;
+    }
+
+    @Override
+    public double jockeyChance() {
+        return level().purpurConfig.drownedJockeyChance;
+    }
+
+    @Override
+    public boolean jockeyTryExistingChickens() {
+        return level().purpurConfig.drownedJockeyTryExistingChickens;
+    }
+    // Purpur end - Configurable jockey options
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.drownedTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.drownedAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
+    // Purpur start - Check mobGriefing Overrides
+    @Override
+    public Boolean checkEntityPickUpLootOverride() {
+        return this.level().purpurConfig.drownedCanPickUpLoot;
+    }
+    // Purpur end - Check mobGriefing Overrides
+
     public static AttributeSupplier.Builder createAttributes() {
         return Zombie.createAttributes().add(Attributes.STEP_HEIGHT, 1.0);
     }
@@ -97,10 +_,23 @@
         this.goalSelector.addGoal(2, new Drowned.DrownedAttackGoal(this, 1.0, false));
         this.goalSelector.addGoal(5, new Drowned.DrownedGoToBeachGoal(this, 1.0));
         this.goalSelector.addGoal(6, new Drowned.DrownedSwimUpGoal(this, 1.0, this.level().getSeaLevel()));
+        if (level().purpurConfig.drownedBreakDoors) this.goalSelector.addGoal(6, new net.minecraft.world.entity.ai.goal.MoveThroughVillageGoal(this, 1.0D, true, 4, this::canBreakDoors)); // Purpur - Option to make drowned break doors
         this.goalSelector.addGoal(7, new RandomStrollGoal(this, 1.0));
         this.targetSelector.addGoal(1, new HurtByTargetGoal(this, Drowned.class).setAlertOthers(ZombifiedPiglin.class));
         this.targetSelector.addGoal(2, new NearestAttackableTargetGoal<>(this, Player.class, 10, true, false, (target, level) -> this.okTarget(target)));
-        this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, AbstractVillager.class, false));
+        // Purpur start - Add option to disable zombie aggressiveness towards villagers
+        if (this.level().spigotConfig.zombieAggressiveTowardsVillager) this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, AbstractVillager.class, false) { // Paper - Check drowned for villager aggression config
+            @Override
+            public boolean canUse() {
+                return (level().purpurConfig.zombieAggressiveTowardsVillagerWhenLagging || !level().getServer().server.isLagging()) && super.canUse();
+            }
+
+            @Override
+            public boolean canContinueToUse() {
+                return (level().purpurConfig.zombieAggressiveTowardsVillagerWhenLagging || !level().getServer().server.isLagging()) && super.canContinueToUse();
+            }
+        });
+        // Purpur end - Add option to disable zombie aggressiveness towards villagers
         this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, IronGolem.class, true));
         this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, Axolotl.class, true, false));
         this.targetSelector.addGoal(5, new NearestAttackableTargetGoal<>(this, Turtle.class, 10, true, false, Turtle.BABY_ON_LAND_SELECTOR));
@@ -239,6 +_,7 @@
 
         LivingEntity target = this.getTarget();
         return target != null && target.isInWater();
+ 
     }
 
     @Override
@@ -279,28 +_,6 @@
         return false;
     }
 
-    private boolean shouldDoRangedAttack() {
-        if (!this.getMainHandItem().is(Items.TRIDENT)) {
-            return false;
-        }
-
-        if (this.hasTargetInRangedDistance()) {
-            return true;
-        }
-
-        LivingEntity target = this.getTarget();
-        if (target == null || !target.isAlive()) {
-            return false;
-        }
-
-        if (this.isWithinMeleeAttackRange(target)) {
-            return false;
-        }
-
-        Path path = this.navigation.getPath() != null ? this.navigation.getPath() : this.navigation.createPath(target, 0);
-        return path != null && path.getTarget().equals(target.blockPosition()) && !path.canReach();
-    }
-
     @Override
     public void performRangedAttack(LivingEntity target, float power) {
         ItemStack mainHandItem = this.getMainHandItem();
@@ -312,7 +_,7 @@
         double distanceToTarget = Math.sqrt(xd * xd + zd * zd);
         if (this.level() instanceof ServerLevel serverLevel) {
             Projectile.spawnProjectileUsingShoot(
-                trident, serverLevel, tridentItemStack, xd, yd + distanceToTarget * 0.2F, zd, 1.6F, this.rangedAttackUncertainty(serverLevel)
+                trident, serverLevel, tridentItemStack, xd, yd + distanceToTarget * 0.2F, zd, 1.6F, 14 - this.level().getDifficulty().getId() * 4
             );
         }
 
@@ -320,16 +_,6 @@
     }
 
     @Override
-    public float rangedAttackUncertainty(Level level) {
-        return this.rangedAttackUncertainty < 0.0F ? RangedAttackMob.super.rangedAttackUncertainty(level) : this.rangedAttackUncertainty;
-    }
-
-    @VisibleForTesting
-    public void setRangedAttackUncertainty(float rangedAttackUncertainty) {
-        this.rangedAttackUncertainty = rangedAttackUncertainty;
-    }
-
-    @Override
     public TagKey<Item> getPreferredWeaponType() {
         return ItemTags.DROWNED_PREFERRED_WEAPONS;
     }
@@ -365,28 +_,12 @@
 
         @Override
         public boolean canUse() {
-            return super.canUse() && this.drowned.okTarget(this.drowned.getTarget()) && !this.drowned.shouldDoRangedAttack();
+            return super.canUse() && this.drowned.okTarget(this.drowned.getTarget());
         }
 
         @Override
         public boolean canContinueToUse() {
-            return super.canContinueToUse() && this.drowned.okTarget(this.drowned.getTarget()) && !this.drowned.shouldDoRangedAttack();
-        }
-
-        @Override
-        public void start() {
-            super.start();
-            if (this.drowned.getMainHandItem().has(DataComponents.WEAPON)) {
-                this.drowned.startUsingItem(InteractionHand.MAIN_HAND);
-            }
-        }
-
-        @Override
-        public void stop() {
-            super.stop();
-            if (this.drowned.getMainHandItem().has(DataComponents.WEAPON)) {
-                this.drowned.stopUsingItem();
-            }
+            return super.canContinueToUse() && this.drowned.okTarget(this.drowned.getTarget());
         }
     }
 
@@ -589,12 +_,7 @@
 
         @Override
         public boolean canUse() {
-            return super.canUse() && this.drowned.shouldDoRangedAttack();
-        }
-
-        @Override
-        public boolean canContinueToUse() {
-            return super.canContinueToUse() && this.drowned.shouldDoRangedAttack();
+            return super.canUse() && this.drowned.getMainHandItem().is(Items.TRIDENT);
         }
 
         @Override
