--- a/net/minecraft/world/entity/monster/zombie/Husk.java
+++ b/net/minecraft/world/entity/monster/zombie/Husk.java
@@ -67,7 +_,7 @@
         boolean result = super.doHurtTarget(level, target);
         if (result && this.getMainHandItem().isEmpty() && target instanceof LivingEntity livingEntity) {
             float difficulty = level.getCurrentDifficultyAt(this.blockPosition()).getEffectiveDifficulty();
-            livingEntity.addEffect(new MobEffectInstance(MobEffects.HUNGER, 140 * (int)difficulty), this);
+            livingEntity.addEffect(new MobEffectInstance(MobEffects.HUNGER, 140 * (int)difficulty), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
         }
 
         return result;
@@ -80,6 +_,7 @@
 
     @Override
     protected void doUnderWaterConversion(ServerLevel level) {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityTypes.ZOMBIE, (timer) -> this.conversionTime = timer)) return;
         this.convertToZombieType(level, EntityTypes.ZOMBIE);
         if (!this.isSilent()) {
             level.levelEvent(null, 1041, this.blockPosition(), 0);
