--- a/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java
+++ b/net/minecraft/world/entity/monster/zombie/ZombifiedPiglin.java
@@ -145,7 +_,7 @@
             .filter(other -> other != this)
             .filter(other -> other.getTarget() == null)
             .filter(other -> !other.isAlliedTo(this.getTarget()))
-            .forEach(other -> other.setTarget(this.getTarget()));
+            .forEach(other -> other.setTarget(this.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true)); // CraftBukkit
     }
 
     private void playAngerSound() {
