--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -66,6 +_,11 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
 import org.jspecify.annotations.Nullable;
 
 public class Zombie extends Monster {
@@ -99,6 +_,7 @@
     private boolean canBreakDoors = false;
     private int inWaterTime = 0;
     public int conversionTime;
+    private boolean shouldBurnInDay = true; // Paper - Add more Zombie API
 
     public Zombie(EntityType<? extends Zombie> type, Level level) {
         super(type, level);
@@ -123,7 +_,7 @@
         this.goalSelector.addGoal(7, new WaterAvoidingRandomStrollGoal(this, 1.0));
         this.targetSelector.addGoal(1, new HurtByTargetGoal(this).setAlertOthers(ZombifiedPiglin.class));
         this.targetSelector.addGoal(2, new NearestAttackableTargetGoal<>(this, Player.class, true));
-        this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, AbstractVillager.class, false));
+        if ( this.level().spigotConfig.zombieAggressiveTowardsVillager ) this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, AbstractVillager.class, false)); // Spigot
         this.targetSelector.addGoal(3, new NearestAttackableTargetGoal<>(this, IronGolem.class, true));
         this.targetSelector.addGoal(5, new NearestAttackableTargetGoal<>(this, Turtle.class, 10, true, false, Turtle.BABY_ON_LAND_SELECTOR));
     }
@@ -232,12 +_,20 @@
         super.tick();
     }
 
+    // Paper start - Add more Zombie API
+    public void stopDrowning() {
+        this.conversionTime = -1;
+        this.getEntityData().set(DATA_DROWNED_CONVERSION_ID, false);
+    }
+    // Paper end - Add more Zombie API
+
     public void startUnderWaterConversion(int time) {
         this.conversionTime = time;
         this.getEntityData().set(DATA_DROWNED_CONVERSION_ID, true);
     }
 
     protected void doUnderWaterConversion(ServerLevel level) {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityTypes.DROWNED, (timer) -> this.conversionTime = timer)) return;
         this.convertToZombieType(level, EntityTypes.DROWNED);
         if (!this.isSilent()) {
             level.levelEvent(null, 1040, this.blockPosition(), 0);
@@ -245,13 +_,19 @@
     }
 
     protected void convertToZombieType(ServerLevel level, EntityType<? extends Zombie> zombieType) {
-        this.convertTo(
+        // CraftBukkit start
+        Zombie converted = this.convertTo(
             zombieType,
             ConversionParams.single(this, true, true),
-            newZombie -> newZombie.handleAttributes(
-                level.getCurrentDifficultyAt(newZombie.blockPosition()).getSpecialMultiplier(), EntitySpawnReason.CONVERSION
-            )
-        );
+            newZombie -> {
+                newZombie.handleAttributes(level.getCurrentDifficultyAt(newZombie.blockPosition()).getSpecialMultiplier(), EntitySpawnReason.CONVERSION);
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, newZombie);
+            // CraftBukkit start
+            }, EntityTransformEvent.TransformReason.DROWNED, CreatureSpawnEvent.SpawnReason.DROWNED);
+        if (converted == null) {
+            ((org.bukkit.entity.Zombie) getBukkitEntity()).setConversionTime(-1); // CraftBukkit - SPIGOT-5208: End conversion to stop event spam
+        }
+        // CraftBukkit end
     }
 
     @VisibleForTesting
@@ -260,7 +_,6 @@
             EntityTypes.ZOMBIE_VILLAGER,
             ConversionParams.single(villager, true, true),
             zombie -> {
-                zombie.setVillagerDataFinalized(villager.getVillagerDataFinalized());
                 zombie.finalizeSpawn(
                     level, level.getCurrentDifficultyAt(zombie.blockPosition()), EntitySpawnReason.CONVERSION, new Zombie.ZombieGroupData(false, true)
                 );
@@ -268,6 +_,7 @@
                 zombie.setGossips(villager.getGossips().copy());
                 zombie.setTradeOffers(villager.getOffers().copy());
                 zombie.setVillagerXp(villager.getVillagerXp());
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(villager, zombie);
                 if (!this.isSilent()) {
                     level.levelEvent(null, 1026, this.blockPosition(), 0);
                 }
@@ -276,9 +_,34 @@
         return zombieVillager != null;
     }
 
+    public static ZombieVillager convertVillagerToZombieVillager(ServerLevel level, Villager villager, net.minecraft.core.BlockPos blockPosition, boolean silent, EntityTransformEvent.TransformReason transformReason, CreatureSpawnEvent.SpawnReason spawnReason) {
+        // CraftBukkit end
+        ZombieVillager zombievillager = (ZombieVillager) villager.convertTo(EntityTypes.ZOMBIE_VILLAGER, ConversionParams.single(villager, true, true), (zombievillager1) -> {
+            zombievillager1.finalizeSpawn(level, level.getCurrentDifficultyAt(zombievillager1.blockPosition()), EntitySpawnReason.CONVERSION, new Zombie.ZombieGroupData(false, true));
+            zombievillager1.setVillagerData(villager.getVillagerData());
+            zombievillager1.setGossips(villager.getGossips().copy());
+            zombievillager1.setTradeOffers(villager.getOffers().copy());
+            zombievillager1.setVillagerXp(villager.getVillagerXp());
+            // CraftBukkit start
+            if (!silent) {
+                level.levelEvent((Entity) null, 1026, blockPosition, 0);
+            }
+
+        }, transformReason, spawnReason);
+
+        return zombievillager;
+        // CraftBukkit end
+    }
+
     public boolean isSunSensitive() {
-        return true;
-    }
+        return this.shouldBurnInDay; // Paper - Add more Zombie API
+    }
+
+    // Paper start - Add more Zombie API
+    public void setShouldBurnInDay(boolean shouldBurnInDay) {
+        this.shouldBurnInDay = shouldBurnInDay;
+    }
+    // Paper end - Add more Zombie API
 
     @Override
     public boolean hurtServer(ServerLevel level, DamageSource source, float damage) {
@@ -342,7 +_,14 @@
         if (result) {
             float difficulty = level.getCurrentDifficultyAt(this.blockPosition()).getEffectiveDifficulty();
             if (this.getMainHandItem().isEmpty() && this.isOnFire() && this.random.nextFloat() < difficulty * 0.3F) {
-                target.igniteForSeconds(2 * (int)difficulty);
+                // CraftBukkit start
+                EntityCombustByEntityEvent event = new EntityCombustByEntityEvent(this.getBukkitEntity(), target.getBukkitEntity(), (float) (2 * (int) difficulty)); // PAIL: fixme
+                this.level().getCraftServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    target.igniteForSeconds(event.getDuration(), false);
+                }
+                // CraftBukkit end
             }
         }
 
@@ -423,7 +_,7 @@
     @Override
     public boolean killedEntity(ServerLevel level, LivingEntity entity, DamageSource source) {
         boolean perished = super.killedEntity(level, entity, source);
-        if ((level.getDifficulty() == Difficulty.NORMAL || level.getDifficulty() == Difficulty.HARD) && entity instanceof Villager villager) {
+        if ((level.getDifficulty() == Difficulty.NORMAL || level.getDifficulty() == Difficulty.HARD) && entity instanceof Villager villager && net.neoforged.neoforge.event.EventHooks.canLivingConvert(entity, EntityTypes.ZOMBIE_VILLAGER, (timer) -> {})) {
             if (level.getDifficulty() != Difficulty.HARD && this.random.nextBoolean()) {
                 return perished;
             }
