--- a/net/minecraft/world/entity/player/ChatVisiblity.java
+++ b/net/minecraft/world/entity/player/ChatVisiblity.java
@@ -1,11 +_,8 @@
 package net.minecraft.world.entity.player;
 
 import com.mojang.serialization.Codec;
-import io.netty.buffer.ByteBuf;
 import java.util.function.IntFunction;
 import net.minecraft.network.chat.Component;
-import net.minecraft.network.codec.ByteBufCodecs;
-import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.util.ByIdMap;
 
 public enum ChatVisiblity {
@@ -15,7 +_,6 @@
 
     private static final IntFunction<ChatVisiblity> BY_ID = ByIdMap.continuous(v -> v.id, values(), ByIdMap.OutOfBoundsStrategy.WRAP);
     public static final Codec<ChatVisiblity> LEGACY_CODEC = Codec.INT.xmap(BY_ID::apply, v -> v.id);
-    public static final StreamCodec<ByteBuf, ChatVisiblity> STREAM_CODEC = ByteBufCodecs.idMapper(BY_ID, v -> v.id);
     private final int id;
     private final Component caption;
 
