--- a/net/minecraft/world/item/trading/Merchant.java
+++ b/net/minecraft/world/item/trading/Merchant.java
@@ -5,10 +_,10 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.world.SimpleMenuProvider;
-import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.inventory.MerchantMenu;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.craftbukkit.inventory.CraftMerchant;
 
 public interface Merchant {
     void setTradingPlayer(@Nullable Player p_45307_);
@@ -49,4 +_,6 @@
     }
 
     boolean isClientSide();
+
+    CraftMerchant getCraftMerchant(); // CraftBukkit
 }
