--- a/net/minecraft/world/item/trading/Merchant.java
+++ b/net/minecraft/world/item/trading/Merchant.java
@@ -5,10 +_,10 @@
 import net.minecraft.network.chat.Component;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.world.SimpleMenuProvider;
-import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.inventory.MerchantMenu;
 import net.minecraft.world.item.ItemStack;
+import org.bukkit.craftbukkit.inventory.CraftMerchant;
 
 public interface Merchant {
     void setTradingPlayer(@Nullable Player p_45307_);
@@ -19,6 +_,7 @@
     MerchantOffers getOffers();
 
     void overrideOffers(MerchantOffers p_45306_);
+    default void processTrade(MerchantOffer merchantRecipe, @Nullable io.papermc.paper.event.player.PlayerPurchaseEvent event) { this.notifyTrade(merchantRecipe); } // Paper
 
     void notifyTrade(MerchantOffer p_45305_);
 
@@ -49,4 +_,6 @@
     }
 
     boolean isClientSide();
+
+    CraftMerchant getCraftMerchant(); // CraftBukkit
 }
