--- a/net/minecraft/world/item/trading/TradeSet.java
+++ b/net/minecraft/world/item/trading/TradeSet.java
@@ -5,18 +_,18 @@
 import java.util.Optional;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.resources.Identifier;
 import net.minecraft.world.level.storage.loot.LootContext;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 
-public record TradeSet(HolderSet<VillagerTrade> trades, Holder<ContextIntProvider> amount, boolean allowDuplicates, Optional<Identifier> randomSequence) {
+public record TradeSet(HolderSet<VillagerTrade> trades, Holder<NumberProvider> amount, boolean allowDuplicates, Optional<Identifier> randomSequence) {
     public static final Codec<TradeSet> CODEC = RecordCodecBuilder.create(
         i -> i.group(
-                RegistryCodecs.holderSet(Registries.VILLAGER_TRADE).fieldOf("trades").forGetter(TradeSet::trades),
-                ContextIntProviders.CODEC.fieldOf("amount").forGetter(TradeSet::amount),
+                RegistryCodecs.homogeneousList(Registries.VILLAGER_TRADE).fieldOf("trades").forGetter(TradeSet::trades),
+                NumberProviders.CODEC.fieldOf("amount").forGetter(TradeSet::amount),
                 Codec.BOOL.optionalFieldOf("allow_duplicates", false).forGetter(TradeSet::allowDuplicates),
                 Identifier.CODEC.optionalFieldOf("random_sequence").forGetter(TradeSet::randomSequence)
             )
