--- a/net/minecraft/world/item/trading/TradeRebalanceVillagerTrades.java
+++ b/net/minecraft/world/item/trading/TradeRebalanceVillagerTrades.java
@@ -14,7 +_,9 @@
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.enchantment.Enchantment;
 import net.minecraft.world.item.enchantment.Enchantments;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
+import net.minecraft.world.level.storage.loot.providers.number.Sum;
+import net.minecraft.world.level.storage.loot.providers.number.UniformGenerator;
 
 public class TradeRebalanceVillagerTrades extends VillagerTrades {
     public static final ResourceKey<VillagerTrade> LIBRARIAN_1_EMERALD_AND_BOOK_DESERT_ENCHANTED_BOOK = resourceKey(
@@ -218,7 +_,7 @@
     public static final ResourceKey<VillagerTrade> ARMORER_5_DIAMOND_BLOCK_EMERALD_TAIGA = resourceKey("armorer/5/diamond_block_emerald_taiga");
     public static final ResourceKey<VillagerTrade> ARMORER_5_IRON_BLOCK_EMERALD_NON_TAIGA = resourceKey("armorer/5/iron_block_emerald_non_taiga");
 
-    public static void bootstrap(BootstrapContext<VillagerTrade> context) {
+    public static Holder<VillagerTrade> bootstrap(BootstrapContext<VillagerTrade> context) {
         HolderGetter<Item> items = context.lookup(Registries.ITEM);
         HolderGetter<VillagerType> villagerVariants = context.lookup(Registries.VILLAGER_TYPE);
         HolderGetter<Enchantment> enchantments = context.lookup(Registries.ENCHANTMENT);
@@ -819,7 +_,7 @@
                 .merchantPredicate(villagerTypeRestriction(villagerTypeHolderSet(villagerVariants, List.of(VillagerType.TAIGA))))
                 .build()
         );
-        register(
+        return register(
             context,
             ARMORER_5_IRON_BLOCK_EMERALD_NON_TAIGA,
             VillagerTrade.builder(new TradeCost(Items.IRON_BLOCK, 1), new ItemStackTemplate(Items.EMERALD, 4), 12, 30, 0.05F)
@@ -869,9 +_,7 @@
         HolderGetter<Item> items, HolderSet<Enchantment> doubleTradePrice, HolderSet<VillagerType> villagerTypes, Holder<Enchantment> enchantment, int level
     ) {
         return VillagerTrade.builder(
-                new TradeCost(
-                    Items.EMERALD, ContextIntProviders.add(ContextIntProviders.exactly(3 * level + 2), ContextIntProviders.between(0, 5 + level * 10))
-                ),
+                new TradeCost(Items.EMERALD, Sum.sum(ConstantValue.exactly(3 * level + 2), UniformGenerator.between(0.0F, 5 + level * 10))),
                 new TradeCost(Items.BOOK, 1),
                 new ItemStackTemplate(Items.ENCHANTED_BOOK),
                 12,
