--- a/net/minecraft/world/item/trading/TradeSets.java
+++ b/net/minecraft/world/item/trading/TradeSets.java
@@ -8,8 +_,8 @@
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.tags.TagKey;
 import net.minecraft.tags.VillagerTradeTags;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
 
 public class TradeSets {
     public static final ResourceKey<TradeSet> WANDERING_TRADER_BUYING = resourceKey("wandering_trader/buying");
@@ -83,7 +_,7 @@
 
     public static Holder<TradeSet> bootstrap(BootstrapContext<TradeSet> context) {
         register(context, WANDERING_TRADER_BUYING, VillagerTradeTags.WANDERING_TRADER_BUYING);
-        register(context, WANDERING_TRADER_COMMON, VillagerTradeTags.WANDERING_TRADER_COMMON, ContextIntProviders.exactly(5));
+        register(context, WANDERING_TRADER_COMMON, VillagerTradeTags.WANDERING_TRADER_COMMON, ConstantValue.exactly(5.0F));
         register(context, WANDERING_TRADER_UNCOMMON, VillagerTradeTags.WANDERING_TRADER_UNCOMMON);
         register(context, ARMORER_LEVEL_1, VillagerTradeTags.ARMORER_LEVEL_1);
         register(context, ARMORER_LEVEL_2, VillagerTradeTags.ARMORER_LEVEL_2);
@@ -129,7 +_,7 @@
         register(context, LIBRARIAN_LEVEL_2, VillagerTradeTags.LIBRARIAN_LEVEL_2);
         register(context, LIBRARIAN_LEVEL_3, VillagerTradeTags.LIBRARIAN_LEVEL_3);
         register(context, LIBRARIAN_LEVEL_4, VillagerTradeTags.LIBRARIAN_LEVEL_4);
-        register(context, LIBRARIAN_LEVEL_5, VillagerTradeTags.LIBRARIAN_LEVEL_5, ContextIntProviders.exactly(3));
+        register(context, LIBRARIAN_LEVEL_5, VillagerTradeTags.LIBRARIAN_LEVEL_5, ConstantValue.exactly(3.0F));
         register(context, MASON_LEVEL_1, VillagerTradeTags.MASON_LEVEL_1);
         register(context, MASON_LEVEL_2, VillagerTradeTags.MASON_LEVEL_2);
         register(context, MASON_LEVEL_3, VillagerTradeTags.MASON_LEVEL_3);
@@ -153,11 +_,11 @@
     }
 
     public static Holder.Reference<TradeSet> register(BootstrapContext<TradeSet> context, ResourceKey<TradeSet> resourceKey, TagKey<VillagerTrade> tradeTag) {
-        return register(context, resourceKey, tradeTag, ContextIntProviders.exactly(2));
+        return register(context, resourceKey, tradeTag, ConstantValue.exactly(2.0F));
     }
 
     public static Holder.Reference<TradeSet> register(
-        BootstrapContext<TradeSet> context, ResourceKey<TradeSet> resourceKey, TagKey<VillagerTrade> tradeTag, Holder<ContextIntProvider> amount
+        BootstrapContext<TradeSet> context, ResourceKey<TradeSet> resourceKey, TagKey<VillagerTrade> tradeTag, Holder<NumberProvider> amount
     ) {
         return context.register(
             resourceKey,
