--- a/net/minecraft/util/debug/DebugBrainDump.java
+++ b/net/minecraft/util/debug/DebugBrainDump.java
@@ -1,6 +_,5 @@
 package net.minecraft.util.debug;
 
-import io.netty.buffer.ByteBuf;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
@@ -15,7 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.GlobalPos;
 import net.minecraft.core.registries.BuiltInRegistries;
-import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.game.DebugEntityNameGenerator;
 import net.minecraft.server.level.ServerLevel;
@@ -53,37 +_,43 @@
     Set<BlockPos> pois,
     Set<BlockPos> potentialPois
 ) {
-    public static final StreamCodec<ByteBuf, DebugBrainDump> STREAM_CODEC = StreamCodec.composite(
-        ByteBufCodecs.STRING_UTF8,
-        DebugBrainDump::name,
-        ByteBufCodecs.STRING_UTF8,
-        DebugBrainDump::profession,
-        ByteBufCodecs.INT,
-        DebugBrainDump::xp,
-        ByteBufCodecs.FLOAT,
-        DebugBrainDump::health,
-        ByteBufCodecs.FLOAT,
-        DebugBrainDump::maxHealth,
-        ByteBufCodecs.STRING_UTF8,
-        DebugBrainDump::inventory,
-        ByteBufCodecs.BOOL,
-        DebugBrainDump::wantsGolem,
-        ByteBufCodecs.INT,
-        DebugBrainDump::angerLevel,
-        ByteBufCodecs.STRING_UTF8.apply(ByteBufCodecs.list()),
-        DebugBrainDump::activities,
-        ByteBufCodecs.STRING_UTF8.apply(ByteBufCodecs.list()),
-        DebugBrainDump::behaviors,
-        ByteBufCodecs.STRING_UTF8.apply(ByteBufCodecs.list()),
-        DebugBrainDump::memories,
-        ByteBufCodecs.STRING_UTF8.apply(ByteBufCodecs.list()),
-        DebugBrainDump::gossips,
-        BlockPos.STREAM_CODEC.apply(ByteBufCodecs.collection(HashSet::new)),
-        DebugBrainDump::pois,
-        BlockPos.STREAM_CODEC.apply(ByteBufCodecs.collection(HashSet::new)),
-        DebugBrainDump::potentialPois,
-        DebugBrainDump::new
-    );
+    public static final StreamCodec<FriendlyByteBuf, DebugBrainDump> STREAM_CODEC = StreamCodec.of((output, value) -> value.write(output), DebugBrainDump::new);
+
+    public DebugBrainDump(FriendlyByteBuf input) {
+        this(
+            input.readUtf(),
+            input.readUtf(),
+            input.readInt(),
+            input.readFloat(),
+            input.readFloat(),
+            input.readUtf(),
+            input.readBoolean(),
+            input.readInt(),
+            input.readList(FriendlyByteBuf::readUtf),
+            input.readList(FriendlyByteBuf::readUtf),
+            input.readList(FriendlyByteBuf::readUtf),
+            input.readList(FriendlyByteBuf::readUtf),
+            input.readCollection(HashSet::new, BlockPos.STREAM_CODEC),
+            input.readCollection(HashSet::new, BlockPos.STREAM_CODEC)
+        );
+    }
+
+    public void write(FriendlyByteBuf output) {
+        output.writeUtf(this.name);
+        output.writeUtf(this.profession);
+        output.writeInt(this.xp);
+        output.writeFloat(this.health);
+        output.writeFloat(this.maxHealth);
+        output.writeUtf(this.inventory);
+        output.writeBoolean(this.wantsGolem);
+        output.writeInt(this.angerLevel);
+        output.writeCollection(this.activities, FriendlyByteBuf::writeUtf);
+        output.writeCollection(this.behaviors, FriendlyByteBuf::writeUtf);
+        output.writeCollection(this.memories, FriendlyByteBuf::writeUtf);
+        output.writeCollection(this.gossips, FriendlyByteBuf::writeUtf);
+        output.writeCollection(this.pois, BlockPos.STREAM_CODEC);
+        output.writeCollection(this.potentialPois, BlockPos.STREAM_CODEC);
+    }
 
     public static DebugBrainDump takeBrainDump(ServerLevel serverLevel, LivingEntity entity) {
         String name = DebugEntityNameGenerator.getEntityName(entity);
