--- a/net/minecraft/util/debug/DebugPathInfo.java
+++ b/net/minecraft/util/debug/DebugPathInfo.java
@@ -7,6 +_,6 @@
 
 public record DebugPathInfo(Path path, float maxNodeDistance) {
     public static final StreamCodec<FriendlyByteBuf, DebugPathInfo> STREAM_CODEC = StreamCodec.composite(
-        Path.DEBUG_STREAM_CODEC, DebugPathInfo::path, ByteBufCodecs.FLOAT, DebugPathInfo::maxNodeDistance, DebugPathInfo::new
+        Path.STREAM_CODEC, DebugPathInfo::path, ByteBufCodecs.FLOAT, DebugPathInfo::maxNodeDistance, DebugPathInfo::new
     );
 }
