--- a/net/minecraft/client/model/effects/SpearAnimations.java
+++ b/net/minecraft/client/model/effects/SpearAnimations.java
@@ -9,6 +_,7 @@
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.util.Ease;
 import net.minecraft.util.Mth;
+import net.minecraft.world.InteractionHand;
 import net.minecraft.world.entity.HumanoidArm;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.component.KineticWeapon;
@@ -21,8 +_,8 @@
         return Mth.clamp(Mth.inverseLerp(time, start, end), 0.0F, 1.0F);
     }
 
-    public static <T extends HumanoidRenderState> void thirdPersonHandUse(ModelPart arm, ModelPart head, HumanoidArm holdingArm, ItemStack item, T state) {
-        int invert = holdingArm == HumanoidArm.RIGHT ? 1 : -1;
+    public static <T extends HumanoidRenderState> void thirdPersonHandUse(ModelPart arm, ModelPart head, boolean holdingInRightArm, ItemStack item, T state) {
+        int invert = holdingInRightArm ? 1 : -1;
         arm.yRot = -0.1F * invert + head.yRot;
         arm.xRot = (float) (-Math.PI / 2) + head.xRot + 0.8F;
         if (state.isFallFlying || state.swimAmount > 0.0F) {
@@ -31,7 +_,8 @@
 
         arm.yRot = (float) (Math.PI / 180.0) * Math.clamp((180.0F / (float)Math.PI) * arm.yRot, -60.0F, 60.0F);
         arm.xRot = (float) (Math.PI / 180.0) * Math.clamp((180.0F / (float)Math.PI) * arm.xRot, -120.0F, 30.0F);
-        if (!(state.ticksUsingItem <= 0.0F) && (!state.isUsingItem || state.useItemHand.asArm(state.mainArm) == holdingArm)) {
+        if (!(state.ticksUsingItem <= 0.0F)
+            && (!state.isUsingItem || state.useItemHand == (holdingInRightArm ? InteractionHand.MAIN_HAND : InteractionHand.OFF_HAND))) {
             KineticWeapon kineticWeapon = item.get(DataComponents.KINETIC_WEAPON);
             if (kineticWeapon != null) {
                 SpearAnimations.UseParams params = SpearAnimations.UseParams.fromKineticWeapon(kineticWeapon, state.ticksUsingItem);
@@ -56,8 +_,8 @@
     ) {
         KineticWeapon kineticWeapon = actualItem.get(DataComponents.KINETIC_WEAPON);
         if (kineticWeapon != null && timeHeld != 0.0F) {
-            float attack = Ease.inQuad(progress(state.swingAnimation, 0.05F, 0.2F));
-            float retract = Ease.inOutExpo(progress(state.swingAnimation, 0.4F, 1.0F));
+            float attack = Ease.inQuad(progress(state.attackTime, 0.05F, 0.2F));
+            float retract = Ease.inOutExpo(progress(state.attackTime, 0.4F, 1.0F));
             SpearAnimations.UseParams params = SpearAnimations.UseParams.fromKineticWeapon(kineticWeapon, timeHeld);
             int invert = arm == HumanoidArm.RIGHT ? 1 : -1;
             float raiseProgressModified = 1.0F - Ease.outBack(1.0F - params.raiseProgress());
@@ -73,24 +_,26 @@
         }
     }
 
-    public static <T extends HumanoidRenderState> void thirdPersonAttackHand(HumanoidModel<T> model, float animation, HumanoidArm arm) {
+    public static <T extends HumanoidRenderState> void thirdPersonAttackHand(HumanoidModel<T> model, T state) {
+        float attackTime = state.attackTime;
+        HumanoidArm arm = state.attackArm;
         model.rightArm.yRot = model.rightArm.yRot - model.body.yRot;
         model.leftArm.yRot = model.leftArm.yRot - model.body.yRot;
         model.leftArm.xRot = model.leftArm.xRot - model.body.yRot;
-        float prepare = Ease.inOutSine(progress(animation, 0.0F, 0.05F));
-        float attack = Ease.inQuad(progress(animation, 0.05F, 0.2F));
-        float retract = Ease.inOutExpo(progress(animation, 0.4F, 1.0F));
+        float prepare = Ease.inOutSine(progress(attackTime, 0.0F, 0.05F));
+        float attack = Ease.inQuad(progress(attackTime, 0.05F, 0.2F));
+        float retract = Ease.inOutExpo(progress(attackTime, 0.4F, 1.0F));
         model.getArm(arm).xRot += (90.0F * prepare - 120.0F * attack + 30.0F * retract) * (float) (Math.PI / 180.0);
     }
 
     public static <S extends ArmedEntityRenderState> void thirdPersonAttackItem(S state, PoseStack poseStack) {
-        float animation = state.swingAnimation;
-        if (!(animation <= 0.0F)) {
+        if (!(state.attackTime <= 0.0F)) {
             KineticWeapon kineticWeapon = state.getMainHandItemStack().get(DataComponents.KINETIC_WEAPON);
             float jetForward = kineticWeapon != null ? kineticWeapon.forwardMovement() : 0.0F;
             float itemInHandDepth = 0.125F;
-            float attack = Ease.inQuad(progress(animation, 0.05F, 0.2F));
-            float retract = Ease.inOutExpo(progress(animation, 0.4F, 1.0F));
+            float attackTime = state.attackTime;
+            float attack = Ease.inQuad(progress(attackTime, 0.05F, 0.2F));
+            float retract = Ease.inOutExpo(progress(attackTime, 0.4F, 1.0F));
             poseStack.rotateAround(Axis.XN.rotationDegrees(70.0F * (attack - retract)), 0.0F, -0.125F, 0.125F);
             poseStack.translate(0.0F, jetForward * (attack - retract), 0.0F);
         }
@@ -141,7 +_,7 @@
         float middleAmount = Ease.outBack(progress(attack, 0.05F, 0.2F));
         float endingAmount = Ease.inOutExpo(progress(attack, 0.4F, 1.0F));
         poseStack.translate(invert * 0.1F * (startingAmount - middleAmount), -0.075F * (startingAmount - endingAmount), 0.65F * (startingAmount - middleAmount));
-        poseStack.rotateDegrees(Axis.XP, -70.0F * (startingAmount - endingAmount));
+        poseStack.mulPose(Axis.XP.rotationDegrees(-70.0F * (startingAmount - endingAmount)));
         poseStack.translate(0.0, 0.0, -0.25 * (endingAmount - middleAmount));
     }
 
