--- a/net/minecraft/client/gui/components/AbstractSliderButton.java
+++ b/net/minecraft/client/gui/components/AbstractSliderButton.java
@@ -30,18 +_,18 @@
     private static final int HANDLE_HALF_WIDTH = 4;
     protected double value;
     protected boolean canChangeValue;
-    protected boolean dragging;
+    private boolean dragging;
 
     public AbstractSliderButton(int x, int y, int width, int height, Component message, double initialValue) {
         super(x, y, width, height, message);
         this.value = initialValue;
     }
 
-    protected Identifier getSprite() {
+    private Identifier getSprite() {
         return this.isActive() && this.isFocused() && !this.canChangeValue ? HIGHLIGHTED_SPRITE : SLIDER_SPRITE;
     }
 
-    protected Identifier getHandleSprite() {
+    private Identifier getHandleSprite() {
         return !this.isActive() || !this.isHovered && !this.canChangeValue ? SLIDER_HANDLE_SPRITE : SLIDER_HANDLE_HIGHLIGHTED_SPRITE;
     }
 
