--- a/net/minecraft/client/gui/components/AbstractSelectionList.java
+++ b/net/minecraft/client/gui/components/AbstractSelectionList.java
@@ -84,7 +_,7 @@
         this.scrollToEntry(this.children.get(secondIndex));
     }
 
-    public void clearEntries() {
+    protected void clearEntries() {
         this.children.clear();
         this.selected = null;
     }
@@ -441,7 +_,7 @@
         private int width = 0;
         private int height;
         @Deprecated
-        protected AbstractSelectionList<E> list;
+        private AbstractSelectionList<E> list;
 
         @Override
         public void setFocused(boolean focused) {
