--- a/net/minecraft/client/gui/components/AbstractWidget.java
+++ b/net/minecraft/client/gui/components/AbstractWidget.java
@@ -92,7 +_,7 @@
 
     protected abstract void extractWidgetRenderState(GuiGraphicsExtractor graphics, int mouseX, int mouseY, float a);
 
-    public void extractScrollingStringOverContents(ActiveTextCollector output, Component message, int margin) {
+    protected void extractScrollingStringOverContents(ActiveTextCollector output, Component message, int margin) {
         int left = this.getX() + margin;
         int right = this.getX() + this.getWidth() - margin;
         int top = this.getY();
