--- a/net/minecraft/client/gui/components/AbstractButton.java
+++ b/net/minecraft/client/gui/components/AbstractButton.java
@@ -18,7 +_,7 @@
 @OnlyIn(Dist.CLIENT)
 public abstract class AbstractButton extends AbstractWidget.WithInactiveMessage {
     protected static final int TEXT_MARGIN = 2;
-    protected static final WidgetSprites SPRITES = new WidgetSprites(
+    private static final WidgetSprites SPRITES = new WidgetSprites(
         Identifier.withDefaultNamespace("widget/button"),
         Identifier.withDefaultNamespace("widget/button_disabled"),
         Identifier.withDefaultNamespace("widget/button_highlighted")
