--- a/net/minecraft/client/gui/components/LogoRenderer.java
+++ b/net/minecraft/client/gui/components/LogoRenderer.java
@@ -15,8 +_,8 @@
     public static final Identifier MINECRAFT_EDITION = Identifier.withDefaultNamespace("textures/gui/title/edition.png");
     public static final int LOGO_WIDTH = 256;
     public static final int LOGO_HEIGHT = 44;
-    public static final int LOGO_TEXTURE_WIDTH = 256;
-    public static final int LOGO_TEXTURE_HEIGHT = 64;
+    private static final int LOGO_TEXTURE_WIDTH = 256;
+    private static final int LOGO_TEXTURE_HEIGHT = 64;
     private static final int EDITION_WIDTH = 128;
     private static final int EDITION_HEIGHT = 14;
     private static final int EDITION_TEXTURE_WIDTH = 128;
