--- a/net/minecraft/client/gui/components/OptionsList.java
+++ b/net/minecraft/client/gui/components/OptionsList.java
@@ -59,7 +_,7 @@
     public void addHeader(Component text) {
         int lineHeight = 9;
         int paddingTop = this.children().isEmpty() ? 0 : lineHeight * 2;
-        this.addEntry(new OptionsList.HeaderEntry(this.screen, text, paddingTop), paddingTop + lineHeight * 2);
+        this.addEntry(new OptionsList.HeaderEntry(this.screen, text, paddingTop), paddingTop + lineHeight + 4);
     }
 
     @Override
@@ -194,15 +_,12 @@
     protected static class HeaderEntry extends OptionsList.AbstractEntry {
         private final Screen screen;
         private final int paddingTop;
-        private final FocusableTextWidget widget;
+        private final StringWidget widget;
 
         protected HeaderEntry(Screen screen, Component text, int paddingTop) {
             this.screen = screen;
             this.paddingTop = paddingTop;
-            this.widget = FocusableTextWidget.builder(text, screen.getFont())
-                .alwaysShowBorder(false)
-                .backgroundFill(FocusableTextWidget.BackgroundFill.ON_FOCUS)
-                .build();
+            this.widget = new StringWidget(text, screen.getFont());
         }
 
         @Override
@@ -212,7 +_,7 @@
 
         @Override
         public void extractContent(GuiGraphicsExtractor graphics, int mouseX, int mouseY, boolean hovered, float a) {
-            this.widget.setPosition(this.screen.width / 2 - this.widget.getWidth() / 2, this.getContentY() + this.paddingTop);
+            this.widget.setPosition(this.screen.width / 2 - 155, this.getContentY() + this.paddingTop);
             this.widget.extractRenderState(graphics, mouseX, mouseY, a);
         }
 
