--- a/net/minecraft/client/gui/components/ScrollableLayout.java
+++ b/net/minecraft/client/gui/components/ScrollableLayout.java
@@ -116,10 +_,6 @@
         return this.container.getHeight();
     }
 
-    public double getScrollAmount() {
-        return this.container.scrollAmount();
-    }
-
     private class Container extends AbstractContainerWidget {
         private final Minecraft minecraft;
         private final List<AbstractWidget> children = new ArrayList<>();
