--- a/net/minecraft/advancements/triggers/EnterBlockTrigger.java
+++ b/net/minecraft/advancements/triggers/EnterBlockTrigger.java
@@ -7,8 +_,8 @@
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockState;
@@ -29,7 +_,7 @@
         public static final Codec<EnterBlockTrigger.TriggerInstance> CODEC = RecordCodecBuilder.create(
             i -> i.group(
                     LootItemCondition.CODEC.optionalFieldOf("player").forGetter(EnterBlockTrigger.TriggerInstance::player),
-                    RegistryCodecs.holderSet(Registries.BLOCK).optionalFieldOf("blocks").forGetter(EnterBlockTrigger.TriggerInstance::block),
+                    RegistryCodecs.homogeneousList(Registries.BLOCK).optionalFieldOf("blocks").forGetter(EnterBlockTrigger.TriggerInstance::block),
                     StatePropertiesPredicate.CODEC.optionalFieldOf("state").forGetter(EnterBlockTrigger.TriggerInstance::state)
                 )
                 .apply(i, EnterBlockTrigger.TriggerInstance::new)
