--- a/net/minecraft/advancements/triggers/FishingRodHookedTrigger.java
+++ b/net/minecraft/advancements/triggers/FishingRodHookedTrigger.java
@@ -60,7 +_,7 @@
 
             if (this.item.isPresent()) {
                 boolean matched = false;
-                Entity hookedInEntity = hookedIn.getOptional(LootContextParams.THIS_ENTITY);
+                Entity hookedInEntity = hookedIn.getOptionalParameter(LootContextParams.THIS_ENTITY);
                 if (hookedInEntity instanceof ItemEntity item && this.item.get().test(item.getItem())) {
                     matched = true;
                 }
