--- a/net/minecraft/client/model/monster/dragon/EnderDragonModel.java
+++ b/net/minecraft/client/model/monster/dragon/EnderDragonModel.java
@@ -112,16 +_,16 @@
         PartDefinition body = root.addOrReplaceChild(
             "body",
             CubeListBuilder.create()
-                .addBox("body", -12.0F, -16.0F, -16.0F, 24, 24, 64, 0, 0)
-                .addBox("scale", -1.0F, -22.0F, -10.0F, 2, 6, 12, 220, 53)
-                .addBox("scale", -1.0F, -22.0F, 10.0F, 2, 6, 12, 220, 53)
-                .addBox("scale", -1.0F, -22.0F, 30.0F, 2, 6, 12, 220, 53),
-            PartPose.offset(0.0F, 20.0F, 8.0F)
+                .addBox("body", -12.0F, 1.0F, -16.0F, 24, 24, 64, 0, 0)
+                .addBox("scale", -1.0F, -5.0F, -10.0F, 2, 6, 12, 220, 53)
+                .addBox("scale", -1.0F, -5.0F, 10.0F, 2, 6, 12, 220, 53)
+                .addBox("scale", -1.0F, -5.0F, 30.0F, 2, 6, 12, 220, 53),
+            PartPose.offset(0.0F, 3.0F, 8.0F)
         );
         PartDefinition leftWing = body.addOrReplaceChild(
             "left_wing",
             CubeListBuilder.create().mirror().addBox("bone", 0.0F, -4.0F, -4.0F, 56, 8, 8, 112, 88).addBox("skin", 0.0F, 0.0F, 2.0F, 56, 0, 56, -56, 88),
-            PartPose.offset(12.0F, -15.0F, -6.0F)
+            PartPose.offset(12.0F, 2.0F, -6.0F)
         );
         leftWing.addOrReplaceChild(
             "left_wing_tip",
@@ -131,7 +_,7 @@
         PartDefinition leftFrontLeg = body.addOrReplaceChild(
             "left_front_leg",
             CubeListBuilder.create().addBox("main", -4.0F, -4.0F, -4.0F, 8, 24, 8, 112, 104),
-            PartPose.offsetAndRotation(12.0F, 0.0F, -6.0F, 1.3F, 0.0F, 0.0F)
+            PartPose.offsetAndRotation(12.0F, 17.0F, -6.0F, 1.3F, 0.0F, 0.0F)
         );
         PartDefinition leftFrontLegTip = leftFrontLeg.addOrReplaceChild(
             "left_front_leg_tip",
@@ -146,7 +_,7 @@
         PartDefinition leftRearLeg = body.addOrReplaceChild(
             "left_hind_leg",
             CubeListBuilder.create().addBox("main", -8.0F, -4.0F, -8.0F, 16, 32, 16, 0, 0),
-            PartPose.offsetAndRotation(16.0F, -4.0F, 34.0F, 1.0F, 0.0F, 0.0F)
+            PartPose.offsetAndRotation(16.0F, 13.0F, 34.0F, 1.0F, 0.0F, 0.0F)
         );
         PartDefinition leftRearLegTip = leftRearLeg.addOrReplaceChild(
             "left_hind_leg_tip",
@@ -161,7 +_,7 @@
         PartDefinition rightWing = body.addOrReplaceChild(
             "right_wing",
             CubeListBuilder.create().addBox("bone", -56.0F, -4.0F, -4.0F, 56, 8, 8, 112, 88).addBox("skin", -56.0F, 0.0F, 2.0F, 56, 0, 56, -56, 88),
-            PartPose.offset(-12.0F, -15.0F, -6.0F)
+            PartPose.offset(-12.0F, 2.0F, -6.0F)
         );
         rightWing.addOrReplaceChild(
             "right_wing_tip",
@@ -171,7 +_,7 @@
         PartDefinition rightFrontLeg = body.addOrReplaceChild(
             "right_front_leg",
             CubeListBuilder.create().addBox("main", -4.0F, -4.0F, -4.0F, 8, 24, 8, 112, 104),
-            PartPose.offsetAndRotation(-12.0F, 0.0F, -6.0F, 1.3F, 0.0F, 0.0F)
+            PartPose.offsetAndRotation(-12.0F, 17.0F, -6.0F, 1.3F, 0.0F, 0.0F)
         );
         PartDefinition rightFrontLegTip = rightFrontLeg.addOrReplaceChild(
             "right_front_leg_tip",
@@ -186,7 +_,7 @@
         PartDefinition rightRearLeg = body.addOrReplaceChild(
             "right_hind_leg",
             CubeListBuilder.create().addBox("main", -8.0F, -4.0F, -8.0F, 16, 32, 16, 0, 0),
-            PartPose.offsetAndRotation(-16.0F, -4.0F, 34.0F, 1.0F, 0.0F, 0.0F)
+            PartPose.offsetAndRotation(-16.0F, 13.0F, 34.0F, 1.0F, 0.0F, 0.0F)
         );
         PartDefinition rightRearLegTip = rightRearLeg.addOrReplaceChild(
             "right_hind_leg_tip",
