--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -1,13 +_,14 @@
 package net.minecraft.server.dedicated;
 
 import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.Lists;
 import com.google.common.collect.ImmutableMap.Builder;
+import com.mohistmc.youer.Metrics;
+import com.mohistmc.youer.WatchMohist;
+import com.mohistmc.youer.YouerConfig;
+import com.mohistmc.youer.util.ProxyUtils;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
-import java.io.BufferedReader;
 import java.io.IOException;
-import java.io.InputStreamReader;
 import java.io.Writer;
 import java.net.InetAddress;
 import java.net.Proxy;
@@ -16,7 +_,7 @@
 import java.nio.file.Files;
 import java.nio.file.LinkOption;
 import java.nio.file.Path;
-import java.util.Collections;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
@@ -35,6 +_,7 @@
 import net.minecraft.server.ServerInterface;
 import net.minecraft.server.ServerLinks;
 import net.minecraft.server.Services;
+import net.minecraft.server.WorldLoader;
 import net.minecraft.server.WorldStem;
 import net.minecraft.server.gui.MinecraftServerGui;
 import net.minecraft.server.jsonrpc.ManagementServer;
@@ -69,6 +_,10 @@
 import net.minecraft.world.level.gamerules.GameRules;
 import net.minecraft.world.level.storage.LevelData;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.io.IoBuilder;
+import org.bukkit.command.CommandSender;
+import org.bukkit.event.server.RemoteServerCommandEvent;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -76,9 +_,9 @@
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int CONVERSION_RETRY_DELAY_MS = 5000;
     private static final int CONVERSION_RETRIES = 2;
-    private final List<ConsoleInput> consoleInput = Collections.synchronizedList(Lists.newArrayList());
+    private final java.util.Queue<ConsoleInput> serverCommandQueue = new java.util.concurrent.ConcurrentLinkedQueue<>(); // Paper - Perf: use a proper queue
     private @Nullable QueryThreadGs4 queryThreadGs4;
-    private final RconConsoleSource rconConsoleSource;
+    // private final RconConsoleSource rconConsoleSource; // CraftBukkit - remove field
     private @Nullable RconThread rconThread;
     public DedicatedServerSettings settings;
     private @Nullable MinecraftServerGui gui;
@@ -87,9 +_,11 @@
     private boolean isTickTimeLoggingEnabled;
     public ServerLinks serverLinks;
     private final Map<String, String> codeOfConductTexts;
-    private final @Nullable ManagementServer jsonRpcServer;
+    private @Nullable ManagementServer jsonRpcServer;
+    private net.neoforged.neoforge.server.dedicated.@Nullable DedicatedLanServerPinger dediLanPinger;
 
     public DedicatedServer(
+        joptsimple.OptionSet options, WorldLoader.DataLoadContext worldLoader,
         Thread serverThread,
         LevelStorageSource.LevelStorageAccess levelStorageSource,
         PackRepository packRepository,
@@ -102,6 +_,7 @@
         NotificationManager notificationManager
     ) {
         super(
+            options, worldLoader,
             serverThread,
             levelStorageSource,
             packRepository,
@@ -115,7 +_,8 @@
             notificationManager
         );
         this.settings = settings;
-        this.rconConsoleSource = new RconConsoleSource(this);
+        this.setMotd(settings.getProperties().motd.get()); // Paper - set field from initial properties
+        // this.rconConsoleSource = new RconConsoleSource(this); // CraftBukkit - remove field
         this.serverTextFilter = ServerTextFilter.createFromConfig(settings.getProperties());
         this.jsonRpcServer = jsonRpcServer;
         this.serverLinks = createServerLinks(settings);
@@ -156,7 +_,7 @@
                 }
             }
 
-            return builder.build();
+            return new LinkedHashMap<>(builder.build()); // CraftBukkit - mutable
         } catch (IOException e) {
             throw new IllegalArgumentException("Failed to read Code of Conduct folder", e);
         }
@@ -167,26 +_,48 @@
         Thread consoleThread = new Thread("Server console handler") {
             @Override
             public void run() {
-                BufferedReader reader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
-
-                String line;
-                try {
-                    while (!DedicatedServer.this.isStopped() && DedicatedServer.this.isRunning() && (line = reader.readLine()) != null) {
-                        DedicatedServer.this.handleConsoleInput(line, DedicatedServer.this.createCommandSourceStack());
-                    }
-                } catch (IOException e) {
-                    DedicatedServer.LOGGER.error("Exception handling console input", e);
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.Main.useConsole) {
+                    return;
                 }
+                if (net.neoforged.neoforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
+                if (System.console() != null) // Purpur - GUI Improvements - has no GUI or has console (did not double-click)
+                new com.destroystokyo.paper.console.PaperConsole(DedicatedServer.this).start();
             }
         };
+
+        // CraftBukkit start - TODO: handle command-line logging arguments
+        java.util.logging.Logger global = java.util.logging.Logger.getLogger("");
+        global.setUseParentHandlers(false);
+        for (java.util.logging.Handler handler : global.getHandlers()) {
+            global.removeHandler(handler);
+        }
+        global.addHandler(new org.bukkit.craftbukkit.util.ForwardLogHandler());
+
+        final org.apache.logging.log4j.Logger logger = LogManager.getRootLogger();
+
+        System.setOut(IoBuilder.forLogger(logger).setLevel(org.apache.logging.log4j.Level.INFO).buildPrintStream());
+        System.setErr(IoBuilder.forLogger(logger).setLevel(org.apache.logging.log4j.Level.WARN).buildPrintStream());
+        // CraftBukkit end
+
         consoleThread.setDaemon(true);
         consoleThread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
-        consoleThread.start();
+        // consoleThread.start(); // Paper - Enhance console tab completions for brigadier commands; moved down
         LOGGER.info("Starting minecraft server version {}", SharedConstants.getCurrentVersion().name());
         if (Runtime.getRuntime().maxMemory() / 1024L / 1024L < 512L) {
             LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
         }
 
+        // Paper start - detect running as root
+        if (io.papermc.paper.util.ServerEnvironment.userIsRootOrAdmin()) {
+            LOGGER.warn("****************************");
+            LOGGER.warn("YOU ARE RUNNING THIS SERVER AS AN ADMINISTRATIVE OR ROOT USER. THIS IS NOT ADVISED.");
+            LOGGER.warn("YOU ARE OPENING YOURSELF UP TO POTENTIAL RISKS WHEN DOING THIS.");
+            LOGGER.warn("FOR MORE INFORMATION, SEE https://madelinemiller.dev/blog/root-minecraft-server/");
+            LOGGER.warn("****************************");
+        }
+        // Paper end - detect running as root
+
         LOGGER.info("Loading properties");
         DedicatedServerProperties properties = this.settings.getProperties();
         if (this.isSingleplayer()) {
@@ -197,57 +_,154 @@
             this.setLocalIp(properties.serverIp);
         }
 
-        this.worldData.setGameType(properties.gameMode.get());
+        // Spigot start
+        this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
+        org.spigotmc.SpigotConfig.init((java.io.File) options.valueOf("spigot-settings"));
+        org.spigotmc.SpigotConfig.registerCommands();
+        // Spigot end
+        // Purpur start - Migrate Setting to reintroduce end void rings
+        try {
+            org.purpurmc.purpur.PurpurConfig.init((java.io.File) options.valueOf("purpur-settings"));
+        } catch (Exception e) {
+            DedicatedServer.LOGGER.error("Unable to load server configuration", e);
+            return false;
+        }
+        org.purpurmc.purpur.PurpurConfig.registerCommands();
+        // Purpur end - Migrate Setting to reintroduce end void rings
+        // Paper start - initialize global and world-defaults configuration
+        this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
+        this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+        // Paper end - initialize global and world-defaults configuration
+        // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
+        if (this.convertOldUsers()) {
+            this.services().nameToIdCache().save(false); // Paper
+        }
+        this.getPlayerList().loadAndSaveFiles(); // Must be after convertNames
+        // Paper end - fix converting txt to json file
+        org.spigotmc.WatchdogThread.doStart(org.spigotmc.SpigotConfig.timeoutTime, org.spigotmc.SpigotConfig.restartOnCrash); // Paper - start watchdog thread
+        consoleThread.start(); // Paper - Enhance console tab completions for brigadier commands; start console thread after MinecraftServer.console & PaperConfig are initialized
+        io.papermc.paper.command.PaperCommands.registerCommands(this); // Paper - setup /paper command
+        com.destroystokyo.paper.Metrics.PaperMetrics.startMetrics(); // Paper - start metrics
+        com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // Paper - load version history now
+
+        // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
-        InetAddress localAddress = null;
-        if (!this.getLocalIp().isEmpty()) {
-            localAddress = InetAddress.getByName(this.getLocalIp());
-        }
+        // Paper start - Unix domain socket support
+        java.net.SocketAddress bindAddress;
+        if (this.getLocalIp().startsWith("unix:")) {
+            if (!io.netty.channel.epoll.Epoll.isAvailable()) {
+                LOGGER.error("**** INVALID CONFIGURATION!");
+                LOGGER.error("You are trying to use a Unix domain socket but you're not on a supported OS.");
+                return false;
+            } else if (!io.papermc.paper.configuration.GlobalConfiguration.get().proxies.velocity.enabled && !org.spigotmc.SpigotConfig.bungee) {
+                LOGGER.error("**** INVALID CONFIGURATION!");
+                LOGGER.error("Unix domain sockets require IPs to be forwarded from a proxy.");
+                return false;
+            }
+            bindAddress = new io.netty.channel.unix.DomainSocketAddress(this.getLocalIp().substring("unix:".length()));
+        } else {
+            InetAddress localAddress = null;
+            if (!this.getLocalIp().isEmpty()) {
+                localAddress = InetAddress.getByName(this.getLocalIp());
+            }
 
-        if (this.getPort() < 0) {
-            this.setPort(properties.serverPort);
+            if (this.getPort() < 0) {
+                this.setPort(properties.serverPort);
+            }
+            bindAddress = new java.net.InetSocketAddress(localAddress, this.getPort());
         }
+        // Paper end - Unix domain socket support
 
         this.initializeKeyPair();
         LOGGER.info("Starting Minecraft server on {}:{}", this.getLocalIp().isEmpty() ? "*" : this.getLocalIp(), this.getPort());
 
         try {
-            this.getConnection().startTcpServerListener(localAddress, this.getPort());
+            this.getConnection().startTcpServerListener(bindAddress); // Paper - Unix domain socket support
         } catch (IOException e) {
             LOGGER.warn("**** FAILED TO BIND TO PORT!");
             LOGGER.warn("The exception was: {}", e.toString());
             LOGGER.warn("Perhaps a server is already running on that port?");
+            if (true) throw new IllegalStateException("Failed to bind to port", e); // Paper - Propagate failed to bind to port error
             return false;
         }
 
+        // Purpur start - UPnP Port Forwarding
+        if (org.purpurmc.purpur.PurpurConfig.useUPnP) {
+            LOGGER.info("[UPnP] Attempting to start UPnP port forwarding service...");
+            if (dev.omega24.upnp4j.UPnP4J.isUPnPAvailable()) {
+                if (dev.omega24.upnp4j.UPnP4J.isOpen(this.getPort(), dev.omega24.upnp4j.util.Protocol.TCP)) {
+                    this.upnp = false;
+                    LOGGER.info("[UPnP] Port {} is already open", this.getPort());
+                } else if (dev.omega24.upnp4j.UPnP4J.open(this.getPort(), dev.omega24.upnp4j.util.Protocol.TCP)) {
+                    this.upnp = true;
+                    LOGGER.info("[UPnP] Successfully opened port {}", this.getPort());
+                } else {
+                    this.upnp = false;
+                    LOGGER.info("[UPnP] Failed to open port {}", this.getPort());
+                }
+
+                if (upnp) {
+                    LOGGER.info("[UPnP] {}:{}", dev.omega24.upnp4j.UPnP4J.getExternalIP(), this.getPort());
+                }
+            } else {
+                this.upnp = false;
+                LOGGER.error("[UPnP] Service is unavailable");
+            }
+        }
+        // Purpur end - UPnP Port Forwarding
+
+        // CraftBukkit start
+        YouerConfig.init((java.io.File) options.valueOf("youer-settings"));
+        YouerConfig.registerCommands();
+        server.loadPlugins();
+        server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
+        // CraftBukkit end
+
+        // Paper start - Add Velocity IP Forwarding Support
+        String proxyFlavor = (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.velocity.enabled) ? "Velocity" : "BungeeCord";
+        String proxyLink = (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.velocity.enabled) ? "https://docs.papermc.io/velocity/security" : "http://www.spigotmc.org/wiki/firewall-guide/";
+        // Paper end - Add Velocity IP Forwarding Support
         if (!this.usesAuthentication()) {
             LOGGER.warn("**** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!");
             LOGGER.warn("The server will make no attempt to authenticate usernames. Beware.");
-            LOGGER.warn(
-                "While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose."
-            );
+            // Spigot start
+            // Paper start - Add Velocity IP Forwarding Support
+            if (ProxyUtils.useProxy()) {
+                LOGGER.warn("Whilst this makes it possible to use {}, unless access to your server is properly restricted, it also opens up the ability for hackers to connect with any username they choose.", proxyFlavor);
+                LOGGER.warn("Please see {} for further information.", proxyLink);
+                // Paper end - Add Velocity IP Forwarding Support
+            } else {
+                LOGGER.warn("While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose.");
+            }
+            // Spigot end
             LOGGER.warn("To change this, set \"online-mode\" to \"true\" in the server.properties file.");
         }
 
+        // CraftBukkit start
+        /*
         if (this.convertOldUsers()) {
             this.services.nameToIdCache().save();
         }
+        */
+        // CraftBukkit end
 
         if (!OldUsersConverter.areOldUserlistsRemoved()) {
             return false;
         }
-
-        this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
+        // this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
         this.tickTimeLogger = new RemoteSampleLogger(TpsDebugDimensions.values().length, this.debugSubscribers(), RemoteDebugSampleType.TICK_TIME);
         long levelNanoTime = Util.getNanos();
         this.services.nameToIdCache().resolveOfflineUsers(!this.usesAuthentication());
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerAboutToStart(this);
         LOGGER.info("Preparing level \"{}\"", this.getLevelIdName());
-        this.loadLevel();
+        this.loadLevel(this.storageSource.getLevelId()); // CraftBukkit
         long elapsed = Util.getNanos() - levelNanoTime;
         String time = String.format(Locale.ROOT, "%.3fs", elapsed / 1.0E9);
-        LOGGER.info("Done ({})! For help, type \"help\"", time);
+        LOGGER.info("Done preparing level \"{}\" ({})", this.getLevelIdName(), time); // Paper - Improve startup message, add total time
+        this.initPostWorld(); // Paper - don't include plugins in world preparation time
+        this.nextTickTimeNanos = Util.getNanos(); // Neo: Update server time to prevent watchdog/spamming during long load.
         if (properties.announcePlayerAchievements != null) {
-            this.getGameRules().set(GameRules.SHOW_ADVANCEMENT_MESSAGES, properties.announcePlayerAchievements, this);
+            this.overworld().getGameRules().set(GameRules.SHOW_ADVANCEMENT_MESSAGES, properties.announcePlayerAchievements, this.overworld()); // CraftBukkit - per-world
         }
 
         if (properties.enableQuery) {
@@ -260,7 +_,7 @@
             this.rconThread = RconThread.create(this);
         }
 
-        if (this.getMaxTickLength() > 0L) {
+        if (false && this.getMaxTickLength() > 0L) {  // Spigot - disable
             Thread watchdog = new Thread(new ServerWatchdog(this), "Server Watchdog");
             watchdog.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(LOGGER));
             watchdog.setDaemon(true);
@@ -272,10 +_,27 @@
             LOGGER.info("JMX monitoring enabled");
         }
 
+        org.purpurmc.purpur.task.BossBarTask.startAll(); // Purpur - Implement TPSBar
+        if (org.purpurmc.purpur.PurpurConfig.beeCountPayload) org.purpurmc.purpur.task.BeehiveTask.instance().register(); // Purpur - Give bee counts in beehives to Purpur clients
+
+        pufferfish$mobSpawnExecutor.start(); // Pufferfish - Async mob spawning
         this.saveEverything(false, true, true);
         this.notificationManager().serverStarted();
+        if (net.neoforged.neoforge.common.config.NeoForgeServerConfig.INSTANCE.advertiseDedicatedServerToLan.get()) {
+            this.dediLanPinger = new net.neoforged.neoforge.server.dedicated.DedicatedLanServerPinger(this.getMotd(), String.valueOf(this.getServerPort()));
+            this.dediLanPinger.start();
+        }
+        Metrics.MohistMetrics.startMetrics();
+        WatchMohist.start();
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
         return true;
     }
+
+    // Paper start
+    public java.io.File getPluginsFolder() {
+        return (java.io.File) this.options.valueOf("plugins");
+    }
+    // Paper end
 
     @Override
     public boolean isEnforceWhitelist() {
@@ -294,6 +_,7 @@
 
     @Override
     public void setUsingWhitelist(boolean usingWhitelist) {
+        new com.destroystokyo.paper.event.server.WhitelistToggleEvent(usingWhitelist).callEvent(); // Paper - WhitelistToggleEvent
         this.settings.update(p -> p.whiteList.update(this.registryAccess(), usingWhitelist));
     }
 
@@ -306,9 +_,9 @@
     }
 
     @Override
-    public boolean saveAllChunks(boolean silent, boolean flush, boolean force) {
+    public boolean saveAllChunks(boolean silent, boolean flush, boolean force, final boolean close) { // Paper - add close param
         this.notificationManager().serverSaveStarted();
-        boolean savedChunks = super.saveAllChunks(silent, flush, force);
+        boolean savedChunks = super.saveAllChunks(silent, flush, force, close); // Paper - add close param
         this.notificationManager().serverSaveCompleted();
         return savedChunks;
     }
@@ -345,7 +_,7 @@
 
     @Override
     protected void forceDifficulty() {
-        this.setDifficulty(this.getProperties().difficulty.get(), true);
+        // this.setDifficulty(this.getProperties().difficulty.get(), true); // Paper - per level difficulty; Don't overwrite level.dat's difficulty, keep current
     }
 
     public int viewDistance() {
@@ -403,11 +_,16 @@
         }
 
         if (this.rconThread != null) {
-            this.rconThread.stop();
+            this.rconThread.stopNonBlocking(); // Paper - don't wait for remote connections
         }
 
         if (this.queryThreadGs4 != null) {
-            this.queryThreadGs4.stop();
+            // this.queryThreadGs4.stop(); // Paper - don't wait for remote connections
+        }
+
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
         }
 
         if (this.jsonRpcServer != null) {
@@ -417,6 +_,8 @@
                 LOGGER.error("Interrupted while stopping the management server", e);
             }
         }
+        this.hasFullyShutdown = true; // Paper - Improved watchdog support
+        System.exit(this.abnormalExit ? 70 : 0); // CraftBukkit // Paper - Improved watchdog support
     }
 
     @Override
@@ -425,13 +_,47 @@
         this.handleConsoleInputs();
     }
 
+    private static final java.util.concurrent.atomic.AtomicInteger ASYNC_DEBUG_CHUNKS_COUNT = new java.util.concurrent.atomic.AtomicInteger(); // Paper - rewrite chunk system
+
     public void handleConsoleInput(String msg, CommandSourceStack source) {
-        this.consoleInput.add(new ConsoleInput(msg, source));
+        // Paper start - rewrite chunk system
+        if (msg.equalsIgnoreCase("paper debug chunks --async")) {
+            LOGGER.info("Scheduling async debug chunks");
+            Runnable run = () -> {
+                LOGGER.info("Async debug chunks executing");
+                ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.dumpAllChunkLoadInfo(this, false);
+                org.bukkit.command.CommandSender sender = MinecraftServer.getServer().console;
+                java.io.File file = ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.getChunkDebugFile();
+                sender.sendMessage(net.kyori.adventure.text.Component.text("Writing chunk information dump to " + file, net.kyori.adventure.text.format.NamedTextColor.GREEN));
+                try {
+                    ca.spottedleaf.moonrise.common.util.JsonUtil.writeJson(ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.debugAllWorlds(this), file);
+                    sender.sendMessage(net.kyori.adventure.text.Component.text("Successfully written chunk information!", net.kyori.adventure.text.format.NamedTextColor.GREEN));
+                } catch (Throwable thr) {
+                    MinecraftServer.LOGGER.warn("Failed to dump chunk information to file " + file.toString(), thr);
+                    sender.sendMessage(net.kyori.adventure.text.Component.text("Failed to dump chunk information, see console", net.kyori.adventure.text.format.NamedTextColor.RED));
+                }
+            };
+            Thread t = new Thread(run);
+            t.setName("Async debug thread #" + ASYNC_DEBUG_CHUNKS_COUNT.getAndIncrement());
+            t.setDaemon(true);
+            t.start();
+            return;
+        }
+        // Paper end - rewrite chunk system
+        this.serverCommandQueue.add(new ConsoleInput(msg, source)); // Paper - Perf: use proper queue
     }
 
     public void handleConsoleInputs() {
-        while (!this.consoleInput.isEmpty()) {
-            ConsoleInput input = this.consoleInput.remove(0);
+        // Paper start - Perf: use proper queue
+        ConsoleInput input;
+        while ((input = this.serverCommandQueue.poll()) != null) {
+            // Paper end - Perf: use proper queue
+            // CraftBukkit start - ServerCommand for preprocessing
+            org.bukkit.event.server.ServerCommandEvent event = new org.bukkit.event.server.ServerCommandEvent(this.console, input.msg);
+            this.server.getPluginManager().callEvent(event);
+            if (event.isCancelled()) continue;
+            input = new ConsoleInput(event.getCommand(), input.source);
+            // CraftBukkit end
             this.getCommands().performPrefixedCommand(input.source, input.msg);
         }
     }
@@ -591,12 +_,15 @@
 
     @Override
     public String getMotd() {
-        return this.settings.getProperties().motd.get();
+        return super.getMotd(); // Paper
     }
 
     @Override
     public void setMotd(String motd) {
-        this.settings.update(p -> p.motd.update(this.registryAccess(), motd));
+        // Paper start
+        super.setMotd(motd);
+        this.settings.update(p -> p.motd.update(this.registryAccess(), this.getMotd()));
+        // Paper end
     }
 
     @Override
@@ -622,7 +_,11 @@
     @Override
     public boolean enforceSecureProfile() {
         DedicatedServerProperties properties = this.getProperties();
-        return properties.enforceSecureProfile && properties.onlineMode && this.services.canValidateProfileKeys();
+        // Paper start - Add setting for proxy online mode status
+        return properties.enforceSecureProfile
+                && io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode()
+                && this.services.canValidateProfileKeys();
+        // Paper end - Add setting for proxy online mode status
     }
 
     @Override
@@ -707,21 +_,65 @@
 
     @Override
     public String getPluginNames() {
-        return "";
+        // CraftBukkit start - Whole method
+        StringBuilder result = new StringBuilder();
+        org.bukkit.plugin.Plugin[] plugins = this.server.getPluginManager().getPlugins();
+
+        result.append(this.server.getName());
+        result.append(" on Bukkit ");
+        result.append(this.server.getBukkitVersion());
+
+        if (plugins.length > 0 && this.server.getQueryPlugins()) {
+            result.append(": ");
+
+            for (int i = 0; i < plugins.length; i++) {
+                if (i > 0) {
+                    result.append("; ");
+                }
+
+                result.append(plugins[i].getDescription().getName());
+                result.append(" ");
+                result.append(plugins[i].getDescription().getVersion().replaceAll(";", ","));
+            }
+        }
+
+        return result.toString();
+        // CraftBukkit end
     }
 
     @Override
     public String runCommand(String command) {
-        this.rconConsoleSource.prepareForCommand();
-        this.executeBlocking(() -> this.getCommands().performPrefixedCommand(this.rconConsoleSource.createCommandSourceStack(), command));
-        return this.rconConsoleSource.getCommandResponse();
+        // CraftBukkit start - fire RemoteServerCommandEvent
+        throw new UnsupportedOperationException("Not supported - remote source required.");
+    }
+
+    public String runCommand(RconConsoleSource rconConsoleSource, String command) {
+        if (command.isBlank()) return ""; // Paper - Do not process empty rcon commands
+
+        rconConsoleSource.prepareForCommand();
+        this.executeBlocking(() -> {
+            CommandSourceStack wrapper = rconConsoleSource.createCommandSourceStack();
+            RemoteServerCommandEvent event = new RemoteServerCommandEvent(rconConsoleSource.getBukkitSender(wrapper), command);
+            server.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            }
+            this.getCommands().performPrefixedCommand(wrapper, event.getCommand());
+        });
+        return rconConsoleSource.getCommandResponse();
+        // CraftBukkit end
     }
 
     @Override
     protected void stopServer() {
         this.notificationManager().serverShuttingDown();
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.GameShuttingDownEvent());
         super.stopServer();
-        Util.shutdownExecutors();
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
+        }
+        //Util.shutdownExecutors(); // Paper - Improved watchdog support; moved into super
     }
 
     @Override
@@ -849,4 +_,15 @@
     public Map<String, String> getCodeOfConducts() {
         return this.codeOfConductTexts;
     }
+
+    // CraftBukkit start
+    public boolean isDebugging() {
+        return this.getProperties().debug;
+    }
+
+    @Override
+    public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+        return console;
+    }
+    // CraftBukkit end
 }
