--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -3,8 +_,12 @@
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
 import com.google.common.collect.ImmutableMap.Builder;
+import com.google.common.net.HostAndPort;
+import com.mohistmc.youer.YouerConfig;
+import com.mohistmc.youer.api.ColorAPI;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
+import io.netty.handler.ssl.SslContext;
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStreamReader;
@@ -17,9 +_,11 @@
 import java.nio.file.LinkOption;
 import java.nio.file.Path;
 import java.util.Collections;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import java.util.stream.Stream;
@@ -35,9 +_,15 @@
 import net.minecraft.server.ServerInterface;
 import net.minecraft.server.ServerLinks;
 import net.minecraft.server.Services;
+import net.minecraft.server.WorldLoader;
 import net.minecraft.server.WorldStem;
 import net.minecraft.server.gui.MinecraftServerGui;
+import net.minecraft.server.jsonrpc.JsonRpcNotificationService;
 import net.minecraft.server.jsonrpc.ManagementServer;
+import net.minecraft.server.jsonrpc.internalapi.MinecraftApi;
+import net.minecraft.server.jsonrpc.security.AuthenticationHandler;
+import net.minecraft.server.jsonrpc.security.JsonRpcSslContextProvider;
+import net.minecraft.server.jsonrpc.security.SecurityConfig;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.level.progress.LoggingLevelLoadListener;
@@ -56,6 +_,7 @@
 import net.minecraft.server.rcon.thread.RconThread;
 import net.minecraft.util.Mth;
 import net.minecraft.util.StringUtil;
+import net.minecraft.util.TimeUtil;
 import net.minecraft.util.Util;
 import net.minecraft.util.debug.DebugSubscriptions;
 import net.minecraft.util.debugchart.RemoteDebugSampleType;
@@ -69,6 +_,17 @@
 import net.minecraft.world.level.gamerules.GameRules;
 import net.minecraft.world.level.storage.LevelData;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.io.IoBuilder;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.util.TerminalConsoleWriterThread;
+import org.bukkit.event.server.RemoteServerCommandEvent;
+import org.bukkit.event.server.ServerCommandEvent;
+import org.jline.reader.EndOfFileException;
+import org.jline.reader.LineReader;
+import org.jline.reader.LineReaderBuilder;
+import org.jline.reader.UserInterruptException;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -78,7 +_,7 @@
     private static final int CONVERSION_RETRIES = 2;
     private final List<ConsoleInput> consoleInput = Collections.synchronizedList(Lists.newArrayList());
     private @Nullable QueryThreadGs4 queryThreadGs4;
-    private final RconConsoleSource rconConsoleSource;
+    // private final RconConsoleSource rconConsoleSource; // CraftBukkit - remove field
     private @Nullable RconThread rconThread;
     public DedicatedServerSettings settings;
     private @Nullable MinecraftServerGui gui;
@@ -87,9 +_,11 @@
     private boolean isTickTimeLoggingEnabled;
     public ServerLinks serverLinks;
     private final Map<String, String> codeOfConductTexts;
-    private final @Nullable ManagementServer jsonRpcServer;
+    private @Nullable ManagementServer jsonRpcServer;
+    private net.neoforged.neoforge.server.dedicated.@Nullable DedicatedLanServerPinger dediLanPinger;
 
     public DedicatedServer(
+        joptsimple.OptionSet options, WorldLoader.DataLoadContext worldLoader,
         Thread serverThread,
         LevelStorageSource.LevelStorageAccess levelStorageSource,
         PackRepository packRepository,
@@ -102,6 +_,7 @@
         NotificationManager notificationManager
     ) {
         super(
+            options, worldLoader,
             serverThread,
             levelStorageSource,
             packRepository,
@@ -115,7 +_,7 @@
             notificationManager
         );
         this.settings = settings;
-        this.rconConsoleSource = new RconConsoleSource(this);
+        // this.rconConsoleSource = new RconConsoleSource(this); // CraftBukkit - remove field
         this.serverTextFilter = ServerTextFilter.createFromConfig(settings.getProperties());
         this.jsonRpcServer = jsonRpcServer;
         this.serverLinks = createServerLinks(settings);
@@ -156,29 +_,99 @@
                 }
             }
 
-            return builder.build();
+            return new LinkedHashMap<>(builder.build()); // CraftBukkit - mutable
         } catch (IOException e) {
             throw new IllegalArgumentException("Failed to read Code of Conduct folder", e);
         }
     }
 
+    private SslContext createSslContext() {
+        try {
+            return JsonRpcSslContextProvider.createFrom(
+                this.getProperties().managementServerTlsKeystore, this.getProperties().managementServerTlsKeystorePassword
+            );
+        } catch (Exception var2) {
+            JsonRpcSslContextProvider.printInstructions();
+            throw new IllegalStateException("Failed to configure TLS for the server management protocol", var2);
+        }
+    }
+
     @Override
     protected boolean initServer() throws IOException {
+        // CraftBukkit start
+        LineReader reader = LineReaderBuilder.builder()
+                .terminal(terminal)
+                .option(LineReader.Option.DISABLE_EVENT_EXPANSION, true)
+                .completer(new org.bukkit.craftbukkit.command.ConsoleCommandCompleter())
+                .build();
+        // CraftBukkit end
         Thread consoleThread = new Thread("Server console handler") {
             @Override
             public void run() {
-                BufferedReader reader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.Main.useConsole) {
+                    return;
+                }
+
+                // MC-33041, SPIGOT-5538: if System.in is not valid due to javaw, then return
+                try {
+                    System.in.available();
+                } catch (IOException ex) {
+                    return;
+                }
+                // CraftBukkit end
+                if (net.neoforged.neoforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
 
                 String line;
                 try {
-                    while (!DedicatedServer.this.isStopped() && DedicatedServer.this.isRunning() && (line = reader.readLine()) != null) {
-                        DedicatedServer.this.handleConsoleInput(line, DedicatedServer.this.createCommandSourceStack());
+                    while (!DedicatedServer.this.isStopped() && DedicatedServer.this.isRunning()) {
+                        try {
+                            if (org.bukkit.craftbukkit.Main.useJline) {
+                                line = reader.readLine(">", null);
+                            } else {
+                                line = reader.readLine();
+                            }
+
+                            if (line.trim().length() > 0) { // Trim to filter lines which are just spaces
+                                DedicatedServer.this.handleConsoleInput(line, DedicatedServer.this.createCommandSourceStack());
+                            }
+                        } catch (EndOfFileException | UserInterruptException eof) {
+                            // SPIGOT-5220: Throttle if EOF (ctrl^d) or stdin is /dev/null
+                            try {
+                                Thread.sleep(50L);
+                            } catch (InterruptedException ex) {
+                                Thread.currentThread().interrupt();
+                            }
+                        }
                     }
-                } catch (IOException e) {
-                    DedicatedServer.LOGGER.error("Exception handling console input", e);
+                } catch (Throwable var4) {
+                    DedicatedServer.LOGGER.error("Exception handling console input", (Throwable) var4);
                 }
             }
         };
+
+        // CraftBukkit start - TODO: handle command-line logging arguments
+        java.util.logging.Logger global = java.util.logging.Logger.getLogger("");
+        global.setUseParentHandlers(false);
+        for (java.util.logging.Handler handler : global.getHandlers()) {
+            global.removeHandler(handler);
+        }
+        global.addHandler(new org.bukkit.craftbukkit.util.ForwardLogHandler());
+
+        final org.apache.logging.log4j.core.Logger logger = ((org.apache.logging.log4j.core.Logger) LogManager.getRootLogger());
+        for (org.apache.logging.log4j.core.Appender appender : logger.getAppenders().values()) {
+            if (appender instanceof org.apache.logging.log4j.core.appender.ConsoleAppender) {
+                logger.removeAppender(appender);
+            }
+        }
+
+        TerminalConsoleWriterThread writerThread = new TerminalConsoleWriterThread(System.out, reader);
+        writerThread.start();
+
+        System.setOut(IoBuilder.forLogger(logger).setLevel(Level.INFO).buildPrintStream());
+        System.setErr(IoBuilder.forLogger(logger).setLevel(Level.WARN).buildPrintStream());
+        // CraftBukkit end
+
         consoleThread.setDaemon(true);
         consoleThread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
         consoleThread.start();
@@ -197,7 +_,13 @@
             this.setLocalIp(properties.serverIp);
         }
 
-        this.worldData.setGameType(properties.gameMode.get());
+        // Spigot start
+        this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
+        org.spigotmc.SpigotConfig.init((java.io.File) options.valueOf("spigot-settings"));
+        org.spigotmc.SpigotConfig.registerCommands();
+        // Spigot end
+
+        // this.worldData.setGameType(properties.gameMode.get()); // CraftBukkit - moved to world loading
         LOGGER.info("Default game type: {}", properties.gameMode.get());
         InetAddress localAddress = null;
         if (!this.getLocalIp().isEmpty()) {
@@ -220,12 +_,24 @@
             return false;
         }
 
+        // CraftBukkit start
+        YouerConfig.init((java.io.File) options.valueOf("youer-settings"));
+        YouerConfig.registerCommands();
+        server.loadPlugins();
+        server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
+        // CraftBukkit end
+
         if (!this.usesAuthentication()) {
             LOGGER.warn("**** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!");
             LOGGER.warn("The server will make no attempt to authenticate usernames. Beware.");
-            LOGGER.warn(
-                "While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose."
-            );
+            // Spigot start
+            if (org.spigotmc.SpigotConfig.bungee) {
+                DedicatedServer.LOGGER.warn("Whilst this makes it possible to use BungeeCord, unless access to your server is properly restricted, it also opens up the ability for hackers to connect with any username they choose.");
+                DedicatedServer.LOGGER.warn("Please see http://www.spigotmc.org/wiki/firewall-guide/ for further information.");
+            } else {
+                DedicatedServer.LOGGER.warn("While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose.");
+            }
+            // Spigot end
             LOGGER.warn("To change this, set \"online-mode\" to \"true\" in the server.properties file.");
         }
 
@@ -236,18 +_,20 @@
         if (!OldUsersConverter.areOldUserlistsRemoved()) {
             return false;
         }
-
-        this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
+        // this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
         this.tickTimeLogger = new RemoteSampleLogger(TpsDebugDimensions.values().length, this.debugSubscribers(), RemoteDebugSampleType.TICK_TIME);
         long levelNanoTime = Util.getNanos();
         this.services.nameToIdCache().resolveOfflineUsers(!this.usesAuthentication());
         LOGGER.info("Preparing level \"{}\"", this.getLevelIdName());
+        youer$loadLevel();
         this.loadLevel();
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerAboutToStart(this);
         long elapsed = Util.getNanos() - levelNanoTime;
         String time = String.format(Locale.ROOT, "%.3fs", elapsed / 1.0E9);
         LOGGER.info("Done ({})! For help, type \"help\"", time);
+        this.nextTickTimeNanos = Util.getNanos(); // Neo: Update server time to prevent watchdog/spamming during long load.
         if (properties.announcePlayerAchievements != null) {
-            this.getGameRules().set(GameRules.SHOW_ADVANCEMENT_MESSAGES, properties.announcePlayerAchievements, this);
+            this.overworld().getGameRules().set(GameRules.SHOW_ADVANCEMENT_MESSAGES, properties.announcePlayerAchievements, this.overworld()); // CraftBukkit - per-world
         }
 
         if (properties.enableQuery) {
@@ -260,7 +_,7 @@
             this.rconThread = RconThread.create(this);
         }
 
-        if (this.getMaxTickLength() > 0L) {
+        if (false && this.getMaxTickLength() > 0L) {  // Spigot - disable
             Thread watchdog = new Thread(new ServerWatchdog(this), "Server Watchdog");
             watchdog.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(LOGGER));
             watchdog.setDaemon(true);
@@ -274,6 +_,12 @@
 
         this.saveEverything(false, true, true);
         this.notificationManager().serverStarted();
+        if (net.neoforged.neoforge.common.config.NeoForgeServerConfig.INSTANCE.advertiseDedicatedServerToLan.get()) {
+            this.dediLanPinger = new net.neoforged.neoforge.server.dedicated.DedicatedLanServerPinger(this.getMotd(), String.valueOf(this.getServerPort()));
+            this.dediLanPinger.start();
+        }
+
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
         return true;
     }
 
@@ -410,6 +_,11 @@
             this.queryThreadGs4.stop();
         }
 
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
+        }
+
         if (this.jsonRpcServer != null) {
             try {
                 this.jsonRpcServer.stop(true);
@@ -417,6 +_,7 @@
                 LOGGER.error("Interrupted while stopping the management server", e);
             }
         }
+        Runtime.getRuntime().halt(0); // CraftBukkit
     }
 
     @Override
@@ -431,8 +_,17 @@
 
     public void handleConsoleInputs() {
         while (!this.consoleInput.isEmpty()) {
-            ConsoleInput input = this.consoleInput.remove(0);
-            this.getCommands().performPrefixedCommand(input.source, input.msg);
+            ConsoleInput consoleinput = (ConsoleInput) this.consoleInput.remove(0);
+
+            // CraftBukkit start - ServerCommand for preprocessing
+            ServerCommandEvent event = new ServerCommandEvent(console, consoleinput.msg);
+            server.getPluginManager().callEvent(event);
+            if (event.isCancelled()) continue;
+            consoleinput = new ConsoleInput(event.getCommand(), consoleinput.source);
+
+            // this.getCommands().performPrefixedCommand(servercommand.source, servercommand.msg); // Called in dispatchServerCommand
+            server.dispatchServerCommand(console, consoleinput);
+            // CraftBukkit end
         }
     }
 
@@ -591,7 +_,7 @@
 
     @Override
     public String getMotd() {
-        return this.settings.getProperties().motd.get();
+        return ColorAPI.string(this.settings.getProperties().motd.get());
     }
 
     @Override
@@ -707,20 +_,63 @@
 
     @Override
     public String getPluginNames() {
-        return "";
+        // CraftBukkit start - Whole method
+        StringBuilder result = new StringBuilder();
+        org.bukkit.plugin.Plugin[] plugins = server.getPluginManager().getPlugins();
+
+        result.append(server.getName());
+        result.append(" on Bukkit ");
+        result.append(server.getBukkitVersion());
+
+        if (plugins.length > 0 && server.getQueryPlugins()) {
+            result.append(": ");
+
+            for (int i = 0; i < plugins.length; i++) {
+                if (i > 0) {
+                    result.append("; ");
+                }
+
+                result.append(plugins[i].getDescription().getName());
+                result.append(" ");
+                result.append(plugins[i].getDescription().getVersion().replaceAll(";", ","));
+            }
+        }
+
+        return result.toString();
+        // CraftBukkit end
     }
 
     @Override
     public String runCommand(String command) {
-        this.rconConsoleSource.prepareForCommand();
-        this.executeBlocking(() -> this.getCommands().performPrefixedCommand(this.rconConsoleSource.createCommandSourceStack(), command));
-        return this.rconConsoleSource.getCommandResponse();
+        // CraftBukkit start - fire RemoteServerCommandEvent
+        throw new UnsupportedOperationException("Not supported - remote source required.");
+    }
+
+    public String runCommand(RconConsoleSource rconConsoleSource, String s) {
+        rconConsoleSource.prepareForCommand();
+        this.executeBlocking(() -> {
+            CommandSourceStack wrapper = rconConsoleSource.createCommandSourceStack();
+            RemoteServerCommandEvent event = new RemoteServerCommandEvent(rconConsoleSource.getBukkitSender(wrapper), s);
+            server.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            }
+            ConsoleInput serverCommand = new ConsoleInput(event.getCommand(), wrapper);
+            server.dispatchServerCommand(event.getSender(), serverCommand);
+        });
+        return rconConsoleSource.getCommandResponse();
+        // CraftBukkit end
     }
 
     @Override
     protected void stopServer() {
         this.notificationManager().serverShuttingDown();
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.GameShuttingDownEvent());
         super.stopServer();
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
+        }
         Util.shutdownExecutors();
     }
 
@@ -849,4 +_,15 @@
     public Map<String, String> getCodeOfConducts() {
         return this.codeOfConductTexts;
     }
+
+    // CraftBukkit start
+    public boolean isDebugging() {
+        return this.getProperties().debug;
+    }
+
+    @Override
+    public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+        return console;
+    }
+    // CraftBukkit end
 }
