--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -1,17 +_,20 @@
 package net.minecraft.server.dedicated;
 
+import com.mohistmc.youer.Metrics;
+import com.mohistmc.youer.YouerConfig;
+import com.mohistmc.youer.WatchMohist;
+import com.mohistmc.youer.api.ServerAPI;
+import com.mohistmc.youer.util.I18n;
 import com.google.common.collect.Lists;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
-import java.io.BufferedReader;
+import io.papermc.paper.console.PaperConsole;
 import java.io.IOException;
-import java.io.InputStreamReader;
 import java.io.Writer;
 import java.net.InetAddress;
 import java.net.Proxy;
 import java.net.URI;
-import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Collections;
@@ -58,6 +_,12 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.SkullBlockEntity;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.io.IoBuilder;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.util.ForwardLogHandler;
+import org.bukkit.event.server.RemoteServerCommandEvent;
+import org.bukkit.event.server.ServerCommandEvent;
 import org.slf4j.Logger;
 
 public class DedicatedServer extends MinecraftServer implements ServerInterface {
@@ -67,7 +_,7 @@
     private final List<ConsoleInput> consoleInput = Collections.synchronizedList(Lists.newArrayList());
     @Nullable
     private QueryThreadGs4 queryThreadGs4;
-    public final RconConsoleSource rconConsoleSource;
+    // public final RconConsoleSource rconConsoleSource;
     @Nullable
     private RconThread rconThread;
     public DedicatedServerSettings settings;
@@ -80,6 +_,8 @@
     @Nullable
     private DebugSampleSubscriptionTracker debugSampleSubscriptionTracker;
     public ServerLinks serverLinks;
+    @Nullable
+    private net.minecraft.client.server.LanServerPinger dediLanPinger;
 
     public DedicatedServer(
         Thread p_214789_,
@@ -93,7 +_,7 @@
     ) {
         super(p_214789_, p_214790_, p_214791_, p_214792_, Proxy.NO_PROXY, p_214794_, p_214795_, p_214796_);
         this.settings = p_214793_;
-        this.rconConsoleSource = new RconConsoleSource(this);
+        // this.rconConsoleSource = new RconConsoleSource(this);
         this.textFilterClient = TextFilterClient.createFromConfig(p_214793_.getProperties().textFilteringConfig);
         this.serverLinks = createServerLinks(p_214793_);
     }
@@ -103,27 +_,36 @@
         Thread thread = new Thread("Server console handler") {
             @Override
             public void run() {
-                BufferedReader bufferedreader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
-
-                String s1;
-                try {
-                    while (!DedicatedServer.this.isStopped() && DedicatedServer.this.isRunning() && (s1 = bufferedreader.readLine()) != null) {
-                        DedicatedServer.this.handleConsoleInput(s1, DedicatedServer.this.createCommandSourceStack());
-                    }
-                } catch (IOException ioexception1) {
-                    DedicatedServer.LOGGER.error("Exception handling console input", (Throwable)ioexception1);
+                if (net.neoforged.neoforge.server.console.TerminalHandler.handleCommands(DedicatedServer.this)) return;
+                if (!org.bukkit.craftbukkit.Main.useConsole) {
+                    return;
                 }
+                new PaperConsole(DedicatedServer.this).start();
             }
         };
+        // CraftBukkit start - TODO: handle command-line logging arguments
+        java.util.logging.Logger global = java.util.logging.Logger.getLogger("");
+        global.setUseParentHandlers(false);
+        for (java.util.logging.Handler handler : global.getHandlers()) {
+            global.removeHandler(handler);
+        }
+        global.addHandler(new ForwardLogHandler());
+
+        // Paper start - Not needed with TerminalConsoleAppender
+        final org.apache.logging.log4j.Logger logger = LogManager.getRootLogger();
+        System.setOut(IoBuilder.forLogger(logger).setLevel(org.apache.logging.log4j.Level.INFO).buildPrintStream());
+        System.setErr(IoBuilder.forLogger(logger).setLevel(org.apache.logging.log4j.Level.WARN).buildPrintStream());
+        // CraftBukkit end
+
         thread.setDaemon(true);
         thread.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandler(LOGGER));
         thread.start();
-        LOGGER.info("Starting minecraft server version {}", SharedConstants.getCurrentVersion().getName());
+        LOGGER.info(I18n.as("minecraftserver.minecraftversion",(Object)SharedConstants.getCurrentVersion().getName()));
         if (Runtime.getRuntime().maxMemory() / 1024L / 1024L < 512L) {
-            LOGGER.warn("To start the server with more ram, launch it as \"java -Xmx1024M -Xms1024M -jar minecraft_server.jar\"");
+            LOGGER.warn(I18n.as("minecraftserver.javamemtoolow"));
         }
 
-        LOGGER.info("Loading properties");
+        LOGGER.info(I18n.as("minecraftserver.loadingproperties"));
         DedicatedServerProperties dedicatedserverproperties = this.settings.getProperties();
         if (this.isSingleplayer()) {
             this.setLocalIp("127.0.0.1");
@@ -138,8 +_,8 @@
         this.setMotd(dedicatedserverproperties.motd);
         super.setPlayerIdleTimeout(dedicatedserverproperties.playerIdleTimeout.get());
         this.setEnforceWhitelist(dedicatedserverproperties.enforceWhitelist);
-        this.worldData.setGameType(dedicatedserverproperties.gamemode);
-        LOGGER.info("Default game type: {}", dedicatedserverproperties.gamemode);
+        // this.worldData.setGameType(dedicatedserverproperties.gamemode); // CraftBukkit - moved to world loading
+        LOGGER.info(I18n.as("minecraftserver.defaultgamemode", dedicatedserverproperties.gamemode));
         InetAddress inetaddress = null;
         if (!this.getLocalIp().isEmpty()) {
             inetaddress = InetAddress.getByName(this.getLocalIp());
@@ -150,24 +_,22 @@
         }
 
         this.initializeKeyPair();
-        LOGGER.info("Starting Minecraft server on {}:{}", this.getLocalIp().isEmpty() ? "*" : this.getLocalIp(), this.getPort());
+        LOGGER.info(I18n.as("minecraftserver.startingserver",this.getLocalIp().isEmpty() ? "*" : this.getLocalIp(), this.getPort()) );
 
         try {
             this.getConnection().startTcpServerListener(inetaddress, this.getPort());
         } catch (IOException ioexception) {
-            LOGGER.warn("**** FAILED TO BIND TO PORT!");
-            LOGGER.warn("The exception was: {}", ioexception.toString());
-            LOGGER.warn("Perhaps a server is already running on that port?");
+            LOGGER.warn(I18n.as("minecraftserver.bindport1"));
+            LOGGER.warn(I18n.as("minecraftserver.bindport2", ioexception.getMessage()));
+            LOGGER.warn(I18n.as("minecraftserver.bindport3",this.getPort()));
             return false;
         }
 
         if (!this.usesAuthentication()) {
-            LOGGER.warn("**** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!");
-            LOGGER.warn("The server will make no attempt to authenticate usernames. Beware.");
-            LOGGER.warn(
-                "While this makes the game possible to play without internet access, it also opens up the ability for hackers to connect with any username they choose."
-            );
-            LOGGER.warn("To change this, set \"online-mode\" to \"true\" in the server.properties file.");
+            LOGGER.warn(I18n.as("minecraftserver.playerauth1"));
+            LOGGER.warn(I18n.as("minecraftserver.playerauth2"));
+            LOGGER.warn(I18n.as("minecraftserver.playerauth3"));
+            LOGGER.warn(I18n.as("minecraftserver.playerauth4"));
         }
 
         if (this.convertOldUsers()) {
@@ -178,6 +_,25 @@
             return false;
         } else {
             this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
+            // Spigot start
+            org.spigotmc.SpigotConfig.init((java.io.File) options.valueOf("spigot-settings"));
+            org.spigotmc.SpigotConfig.registerCommands();
+            // Spigot end
+            // Paper start - initialize global and world-defaults configuration
+            this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
+            this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+            // Paper end - initialize global and world-defaults configuration
+            // Mohsit start
+            YouerConfig.init((java.io.File) options.valueOf("youer-settings"));
+            YouerConfig.registerCommands();
+            if (YouerConfig.motdEnable) {
+                this.setMotd(YouerConfig.motd());
+            }
+            // Youer end
+            // CraftBukkit start
+            server.loadPlugins();
+            server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.STARTUP);
+            // CraftBukkit end
             this.debugSampleSubscriptionTracker = new DebugSampleSubscriptionTracker(this.getPlayerList());
             this.tickTimeLogger = new RemoteSampleLogger(
                 TpsDebugDimensions.values().length, this.debugSampleSubscriptionTracker, RemoteDebugSampleType.TICK_TIME
@@ -185,11 +_,13 @@
             long i = Util.getNanos();
             SkullBlockEntity.setup(this.services, this);
             GameProfileCache.setUsesAuthentication(this.usesAuthentication());
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerAboutToStart(this);
             LOGGER.info("Preparing level \"{}\"", this.getLevelIdName());
             this.loadLevel();
             long j = Util.getNanos() - i;
             String s = String.format(Locale.ROOT, "%.3fs", (double)j / 1.0E9);
-            LOGGER.info("Done ({})! For help, type \"help\"", s);
+            LOGGER.info(I18n.as("minecraftserver.startdone", s));
+            this.nextTickTimeNanos = Util.getNanos(); // Neo: Update server time to prevent watchdog/spaming during long load.
             if (dedicatedserverproperties.announcePlayerAchievements != null) {
                 this.getGameRules().getRule(GameRules.RULE_ANNOUNCE_ADVANCEMENTS).set(dedicatedserverproperties.announcePlayerAchievements, this);
             }
@@ -204,7 +_,7 @@
                 this.rconThread = RconThread.create(this);
             }
 
-            if (this.getMaxTickLength() > 0L) {
+            if (false && this.getMaxTickLength() > 0L) { // Spigot - disable
                 Thread thread1 = new Thread(new ServerWatchdog(this));
                 thread1.setUncaughtExceptionHandler(new DefaultUncaughtExceptionHandlerWithName(LOGGER));
                 thread1.setName("Server Watchdog");
@@ -217,10 +_,24 @@
                 LOGGER.info("JMX monitoring enabled");
             }
 
+            if (net.neoforged.neoforge.common.NeoForgeConfig.SERVER.advertiseDedicatedServerToLan.get()) {
+                this.dediLanPinger = new net.minecraft.client.server.LanServerPinger(this.getMotd(), String.valueOf(this.getServerPort()));
+                this.dediLanPinger.start();
+            }
+            Metrics.MohistMetrics.startMetrics();
+            WatchMohist.start();
+            org.spigotmc.WatchdogThread.doStart(org.spigotmc.SpigotConfig.timeoutTime, org.spigotmc.SpigotConfig.restartOnCrash); // Paper - start watchdog thread
+            net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
             return true;
         }
     }
 
+    // Paper start
+    public java.io.File getPluginsFolder() {
+        return (java.io.File) this.options.valueOf("plugins");
+    }
+    // Paper end
+
     @Override
     public boolean isSpawningAnimals() {
         return this.getProperties().spawnAnimals && super.isSpawningAnimals();
@@ -295,6 +_,16 @@
         if (this.queryThreadGs4 != null) {
             this.queryThreadGs4.stop();
         }
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
+        }
+
+        if (ServerAPI.yes_steve_model()) {
+            System.exit(0);
+        } else {
+            Runtime.getRuntime().halt(0);
+        }
     }
 
     @Override
@@ -315,7 +_,15 @@
     public void handleConsoleInputs() {
         while (!this.consoleInput.isEmpty()) {
             ConsoleInput consoleinput = this.consoleInput.remove(0);
-            this.getCommands().performPrefixedCommand(consoleinput.source, consoleinput.msg);
+            // CraftBukkit start - ServerCommand for preprocessing
+            ServerCommandEvent event = new ServerCommandEvent(console, consoleinput.msg);
+            server.getPluginManager().callEvent(event);
+            if (event.isCancelled()) continue;
+            consoleinput = new ConsoleInput(event.getCommand(), consoleinput.source);
+
+            // this.getCommands().performCommand(servercommand.source, servercommand.msg); // Called in dispatchServerCommand
+            server.dispatchServerCommand(console, consoleinput);
+            // CraftBukkit end
         }
     }
 
@@ -532,14 +_,51 @@
 
     @Override
     public String getPluginNames() {
-        return "";
+        // CraftBukkit start - Whole method
+        StringBuilder result = new StringBuilder();
+        org.bukkit.plugin.Plugin[] plugins = this.server.getPluginManager().getPlugins();
+
+        result.append(this.server.getName());
+        result.append(" on Bukkit ");
+        result.append(this.server.getBukkitVersion());
+
+        if (plugins.length > 0 && this.server.getQueryPlugins()) {
+            result.append(": ");
+
+            for (int i = 0; i < plugins.length; i++) {
+                if (i > 0) {
+                    result.append("; ");
+                }
+
+                result.append(plugins[i].getDescription().getName());
+                result.append(" ");
+                result.append(plugins[i].getDescription().getVersion().replaceAll(";", ","));
+            }
+        }
+
+        return result.toString();
+        // CraftBukkit end
     }
 
     @Override
     public String runCommand(String p_139644_) {
-        this.rconConsoleSource.prepareForCommand();
-        this.executeBlocking(() -> this.getCommands().performPrefixedCommand(this.rconConsoleSource.createCommandSourceStack(), p_139644_));
-        return this.rconConsoleSource.getCommandResponse();
+        throw new UnsupportedOperationException("Not supported - remote source required.");
+    }
+
+    public String runCommand(RconConsoleSource rconConsoleSource, String s) {
+        rconConsoleSource.prepareForCommand();
+        this.executeBlocking(() -> {
+            CommandSourceStack wrapper = rconConsoleSource.createCommandSourceStack();
+            RemoteServerCommandEvent event = new RemoteServerCommandEvent(rconConsoleSource.getBukkitSender(wrapper), s);
+            server.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            }
+            ConsoleInput serverCommand = new ConsoleInput(event.getCommand(), wrapper);
+            server.dispatchServerCommand(event.getSender(), serverCommand);
+        });
+        return rconConsoleSource.getCommandResponse();
+        // CraftBukkit end
     }
 
     public void storeUsingWhiteList(boolean p_139689_) {
@@ -548,7 +_,12 @@
 
     @Override
     public void stopServer() {
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.GameShuttingDownEvent());
         super.stopServer();
+        if (this.dediLanPinger != null) {
+            this.dediLanPinger.interrupt();
+            this.dediLanPinger = null;
+        }
         Util.shutdownExecutors();
         SkullBlockEntity.clear();
     }
@@ -638,5 +_,15 @@
                 return Optional.empty();
             }
         }
+    }
+
+    @Override
+    public boolean isDebugging() {
+        return this.getProperties().debug;
+    }
+
+    @Override
+    public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+        return console;
     }
 }
