--- a/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ b/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -18,6 +_,11 @@
         this.setViewDistance(dedicatedserverproperties.viewDistance);
         this.setSimulationDistance(dedicatedserverproperties.simulationDistance);
         super.setUsingWhiteList(dedicatedserverproperties.whiteList.get());
+        // Paper start - fix converting txt to json file; moved from constructor
+    }
+    @Override
+    public void loadAndSaveFiles() {
+        // Paper end - fix converting txt to json file
         this.loadUserBanList();
         this.saveUserBanList();
         this.loadIpBanList();
