--- a/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java
+++ b/net/minecraft/network/protocol/common/ClientboundUpdateTagsPacket.java
@@ -1,23 +_,31 @@
 package net.minecraft.network.protocol.common;
 
-import io.netty.buffer.ByteBuf;
-import java.util.IdentityHashMap;
 import java.util.Map;
 import net.minecraft.core.Registry;
-import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.tags.TagNetworkSerialization;
 
-public record ClientboundUpdateTagsPacket(Map<ResourceKey<? extends Registry<?>>, TagNetworkSerialization.NetworkPayload> tags)
-    implements Packet<ClientCommonPacketListener> {
-    public static final StreamCodec<ByteBuf, ClientboundUpdateTagsPacket> STREAM_CODEC = StreamCodec.composite(
-        ByteBufCodecs.map(IdentityHashMap::new, ResourceKey.REGISTRY_STREAM_CODEC, TagNetworkSerialization.NetworkPayload.STREAM_CODEC),
-        ClientboundUpdateTagsPacket::tags,
-        ClientboundUpdateTagsPacket::new
+public class ClientboundUpdateTagsPacket implements Packet<ClientCommonPacketListener> {
+    public static final StreamCodec<FriendlyByteBuf, ClientboundUpdateTagsPacket> STREAM_CODEC = Packet.codec(
+        ClientboundUpdateTagsPacket::write, ClientboundUpdateTagsPacket::new
     );
+    private final Map<ResourceKey<? extends Registry<?>>, TagNetworkSerialization.NetworkPayload> tags;
+
+    public ClientboundUpdateTagsPacket(Map<ResourceKey<? extends Registry<?>>, TagNetworkSerialization.NetworkPayload> tags) {
+        this.tags = tags;
+    }
+
+    private ClientboundUpdateTagsPacket(FriendlyByteBuf input) {
+        this.tags = input.readMap(FriendlyByteBuf::readRegistryKey, TagNetworkSerialization.NetworkPayload::read);
+    }
+
+    private void write(FriendlyByteBuf output) {
+        output.writeMap(this.tags, FriendlyByteBuf::writeResourceKey, (buffer, value) -> value.write(buffer));
+    }
 
     @Override
     public PacketType<ClientboundUpdateTagsPacket> type() {
@@ -26,5 +_,9 @@
 
     public void handle(ClientCommonPacketListener listener) {
         listener.handleUpdateTags(this);
+    }
+
+    public Map<ResourceKey<? extends Registry<?>>, TagNetworkSerialization.NetworkPayload> getTags() {
+        return this.tags;
     }
 }
