--- a/net/minecraft/client/model/animal/golem/CopperGolemModel.java
+++ b/net/minecraft/client/model/animal/golem/CopperGolemModel.java
@@ -287,7 +_,7 @@
         ModelPart activeArm = arm == HumanoidArm.RIGHT ? this.rightArm : this.leftArm;
         activeArm.translateAndRotate(poseStack);
         if (state.copperGolemState.equals(CopperGolemState.IDLE)) {
-            poseStack.rotateDegrees(Axis.YP, arm == HumanoidArm.RIGHT ? -90.0F : 90.0F);
+            poseStack.mulPose(Axis.YP.rotationDegrees(arm == HumanoidArm.RIGHT ? -90.0F : 90.0F));
             poseStack.translate(0.0F, 0.0F, 0.125F);
         } else {
             poseStack.scale(0.55F, 0.55F, 0.55F);
