--- a/net/minecraft/client/Camera.java
+++ b/net/minecraft/client/Camera.java
@@ -120,7 +_,6 @@
         cameraState.initialized = this.isInitialized();
         cameraState.isPanoramicMode = this.isPanoramicMode;
         cameraState.isFrustumCaptured = this.capturedFrustum != null;
-        cameraState.isFirstPerson = this.minecraft.options.getCameraType().isFirstPerson();
         cameraState.smartCull = this.minecraft.smartCull;
         if (this.minecraft.player.isSpectator() && this.level.getBlockState(this.blockPosition).isSolidRender()) {
             cameraState.smartCull = false;
