--- a/net/minecraft/client/KeyMapping.java
+++ b/net/minecraft/client/KeyMapping.java
@@ -165,15 +_,15 @@
     }
 
     public boolean matches(KeyEvent event) {
-        return this.key.getType() == InputConstants.Type.KEYBOARD && this.key.getValue() == event.key() && !this.isUnbound();
+        return this.key.getType() == InputConstants.Type.KEYBOARD && this.key.getValue() == event.key();
     }
 
     public boolean matchesMouse(MouseButtonEvent event) {
-        return this.key.getType() == InputConstants.Type.MOUSE && this.key.getValue() == event.button() && !this.isUnbound();
+        return this.key.getType() == InputConstants.Type.MOUSE && this.key.getValue() == event.button();
     }
 
     public boolean matches(InputConstants.Key key) {
-        return this.key.equals(key) && !this.isUnbound();
+        return this.key.equals(key);
     }
 
     public Component getTranslatedKeyMessage() {
