--- a/net/minecraft/client/KeyMapping.java
+++ b/net/minecraft/client/KeyMapping.java
@@ -18,17 +_,15 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.extensions.IKeyMappingExtension;
-
 @OnlyIn(Dist.CLIENT)
-public class KeyMapping implements Comparable<KeyMapping>, IKeyMappingExtension {
+public class KeyMapping implements Comparable<KeyMapping> {
     private static final Map<String, KeyMapping> ALL = Maps.newHashMap();
     private static final Map<InputConstants.Key, List<KeyMapping>> MAP = Maps.newHashMap();
     private final String name;
     private final InputConstants.Key defaultKey;
     private final KeyMapping.Category category;
     protected InputConstants.Key key;
-    boolean isDown;
+    private boolean isDown;
     private int clickCount;
     private final int order;
 
@@ -205,7 +_,7 @@
     }
 
     public record Category(Identifier id) {
-        static final List<KeyMapping.Category> SORT_ORDER = new ArrayList<>();
+        private static final List<KeyMapping.Category> SORT_ORDER = new ArrayList<>();
         public static final KeyMapping.Category MOVEMENT = register("movement");
         public static final KeyMapping.Category MISC = register("misc");
         public static final KeyMapping.Category MULTIPLAYER = register("multiplayer");
