--- a/net/minecraft/client/Minecraft.java
+++ b/net/minecraft/client/Minecraft.java
@@ -256,10 +_,8 @@
 import org.lwjgl.glfw.GLFW;
 import org.slf4j.Logger;
 
-import net.neoforged.neoforge.client.extensions.IMinecraftExtension;
-
 @OnlyIn(Dist.CLIENT)
-public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements WindowEventHandler, IMinecraftExtension {
+public class Minecraft extends ReentrantBlockableEventLoop<Runnable> implements WindowEventHandler {
     private static Minecraft instance;
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int MAX_TICKS_PER_UPDATE = 10;
