--- a/net/minecraft/client/Minecraft.java
+++ b/net/minecraft/client/Minecraft.java
@@ -373,7 +_,7 @@
     private boolean surfaceIsInvalid = true;
     private @Nullable BackendCreationException backendCreationException;
 
-    public Minecraft(GameConfig gameConfig) {
+    public Minecraft(GameConfig gameConfig) throws Exception {
         super("Client", true);
         instance = this;
         this.clientStartTimeMs = System.currentTimeMillis();
@@ -395,7 +_,7 @@
         YggdrasilAuthenticationService authenticationService = this.offlineDeveloperMode
             ? YggdrasilAuthenticationService.createOffline(this.proxy)
             : new YggdrasilAuthenticationService(this.proxy);
-        this.services = Services.create(authenticationService, this.gameDirectory);
+        this.services = Services.create(authenticationService, this.gameDirectory, null, null); // Youer
         this.user = gameConfig.user.user;
         this.profileFuture = this.offlineDeveloperMode
             ? CompletableFuture.completedFuture(null)
