--- a/net/minecraft/client/MouseHandler.java
+++ b/net/minecraft/client/MouseHandler.java
@@ -4,8 +_,6 @@
 import com.mojang.blaze3d.platform.InputConstants;
 import com.mojang.blaze3d.platform.Window;
 import com.mojang.logging.LogUtils;
-import java.nio.FloatBuffer;
-import java.nio.IntBuffer;
 import java.nio.file.InvalidPathException;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -30,9 +_,6 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.joml.Vector2i;
 import org.jspecify.annotations.Nullable;
-import org.lwjgl.sdl.SDLMouse;
-import org.lwjgl.sdl.SDLVideo;
-import org.lwjgl.system.MemoryStack;
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
@@ -47,6 +_,7 @@
     private double ypos;
     private MouseHandler.@Nullable LastClick lastClick;
     protected @MouseButtonInfo.MouseButton int lastClickButton;
+    private int fakeRightMouse;
     private @Nullable MouseButtonInfo activeButton = null;
     private boolean ignoreFirstMove = true;
     private double mousePressedTime;
@@ -72,8 +_,9 @@
             }
 
             boolean pressed = action == 1;
+            MouseButtonInfo buttonInfo = this.simulateRightClick(rawButtonInfo, pressed);
             if (pressed) {
-                this.activeButton = rawButtonInfo;
+                this.activeButton = buttonInfo;
                 this.mousePressedTime = Blaze3D.getTime();
             } else if (this.activeButton != null) {
                 this.activeButton = null;
@@ -81,8 +_,7 @@
 
             if (this.minecraft.gui.overlay() == null) {
                 if (pressed
-                    && this.minecraft
-                        .handleGlobalKeyPress(InputConstants.Type.MOUSE.getOrCreate(rawButtonInfo.button()), rawButtonInfo.hasControlDownWithQuirk())) {
+                    && this.minecraft.handleGlobalKeyPress(InputConstants.Type.MOUSE.getOrCreate(buttonInfo.button()), buttonInfo.hasControlDownWithQuirk())) {
                     return;
                 }
 
@@ -94,7 +_,7 @@
                     double xm = this.getScaledXPos(window);
                     double ym = this.getScaledYPos(window);
                     Screen screen = this.minecraft.gui.screen();
-                    MouseButtonEvent event = new MouseButtonEvent(xm, ym, rawButtonInfo);
+                    MouseButtonEvent event = new MouseButtonEvent(xm, ym, buttonInfo);
                     if (pressed) {
                         screen.afterMouseAction();
 
@@ -106,11 +_,12 @@
                                 && this.lastClickButton == event.button();
                             if (screen.mouseClicked(event, doubleClick)) {
                                 this.lastClick = new MouseHandler.LastClick(currentTime, screen);
-                                this.lastClickButton = rawButtonInfo.button();
+                                this.lastClickButton = buttonInfo.button();
                                 return;
                             }
                         } catch (Throwable t) {
                             CrashReport report = CrashReport.forThrowable(t, "mouseClicked event handler");
+                            screen.fillCrashDetails(report);
                             CrashReportCategory mouseDetails = report.addCategory("Mouse");
                             this.fillMousePositionDetails(mouseDetails, window);
                             mouseDetails.setDetail("Button", event.button());
@@ -123,6 +_,7 @@
                             }
                         } catch (Throwable t) {
                             CrashReport report = CrashReport.forThrowable(t, "mouseReleased event handler");
+                            screen.fillCrashDetails(report);
                             CrashReportCategory mouseDetails = report.addCategory("Mouse");
                             this.fillMousePositionDetails(mouseDetails, window);
                             mouseDetails.setDetail("Button", event.button());
@@ -133,15 +_,15 @@
             }
 
             if (this.minecraft.gui.screen() == null && this.minecraft.gui.overlay() == null) {
-                if (rawButtonInfo.button() == 1) {
+                if (buttonInfo.button() == 1) {
                     this.isLeftPressed = pressed;
-                } else if (rawButtonInfo.button() == 2) {
+                } else if (buttonInfo.button() == 2) {
                     this.isMiddlePressed = pressed;
-                } else if (rawButtonInfo.button() == 3) {
+                } else if (buttonInfo.button() == 3) {
                     this.isRightPressed = pressed;
                 }
 
-                InputConstants.Key mouseKey = InputConstants.Type.MOUSE.getOrCreate(rawButtonInfo.button());
+                InputConstants.Key mouseKey = InputConstants.Type.MOUSE.getOrCreate(buttonInfo.button());
                 KeyMapping.set(mouseKey, pressed);
                 if (pressed) {
                     KeyMapping.click(mouseKey);
@@ -150,6 +_,22 @@
         }
     }
 
+    private MouseButtonInfo simulateRightClick(MouseButtonInfo info, boolean pressed) {
+        if (InputQuirks.SIMULATE_RIGHT_CLICK_WITH_LONG_LEFT_CLICK && info.button() == 1) {
+            if (pressed) {
+                if ((info.modifiers() & 192) == 192) {
+                    this.fakeRightMouse++;
+                    return new MouseButtonInfo(3, info.modifiers());
+                }
+            } else if (this.fakeRightMouse > 0) {
+                this.fakeRightMouse--;
+                return new MouseButtonInfo(3, info.modifiers());
+            }
+        }
+
+        return info;
+    }
+
     public void fillMousePositionDetails(CrashReportCategory category, Window window) {
         category.setDetail(
             "Mouse location",
@@ -275,6 +_,7 @@
                     screen.mouseMoved(xm, ym);
                 } catch (Throwable t) {
                     CrashReport report = CrashReport.forThrowable(t, "mouseMoved event handler");
+                    screen.fillCrashDetails(report);
                     CrashReportCategory mouseDetails = report.addCategory("Mouse");
                     this.fillMousePositionDetails(mouseDetails, window);
                     throw new ReportedException(report);
@@ -288,6 +_,7 @@
                         screen.mouseDragged(new MouseButtonEvent(xm, ym, this.activeButton), dx, dy);
                     } catch (Throwable t) {
                         CrashReport report = CrashReport.forThrowable(t, "mouseDragged event handler");
+                        screen.fillCrashDetails(report);
                         CrashReportCategory mouseDetails = report.addCategory("Mouse");
                         this.fillMousePositionDetails(mouseDetails, window);
                         throw new ReportedException(report);
@@ -373,26 +_,6 @@
 
     public void setIgnoreFirstMove() {
         this.ignoreFirstMove = true;
-    }
-
-    public void resyncMousePosition() {
-        if (!this.mouseGrabbed) {
-            try (MemoryStack stack = MemoryStack.stackPush()) {
-                FloatBuffer x = stack.mallocFloat(1);
-                FloatBuffer y = stack.mallocFloat(1);
-                IntBuffer windowX = stack.mallocInt(1);
-                IntBuffer windowY = stack.mallocInt(1);
-                SDLMouse.SDL_GetGlobalMouseState(x, y);
-                if (SDLVideo.SDL_GetWindowPosition(this.minecraft.getWindow().handle(), windowX, windowY)) {
-                    this.xpos = x.get(0) - windowX.get(0);
-                    this.ypos = y.get(0) - windowY.get(0);
-                } else {
-                    SDLMouse.SDL_GetMouseState(x, y);
-                    this.xpos = x.get(0);
-                    this.ypos = y.get(0);
-                }
-            }
-        }
     }
 
     public boolean isMouseGrabbed() {
