--- a/net/minecraft/client/OptionInstance.java
+++ b/net/minecraft/client/OptionInstance.java
@@ -42,12 +_,12 @@
         : CommonComponents.OPTION_OFF;
     public static final OptionInstance.ValueUpdateListener<Object> NO_ACTION = var0 -> {};
     private final OptionInstance.TooltipSupplier<T> tooltip;
-    public final Function<T, Component> toString;
+    private final Function<T, Component> toString;
     private final OptionInstance.ValueSet<T> values;
     private final Codec<T> codec;
     private final T initialValue;
     private final OptionInstance.ValueUpdateListener<? super T> onValueUpdate;
-    public final Component caption;
+    private final Component caption;
     private T value;
 
     public static OptionInstance<Boolean> createBoolean(
