--- a/net/minecraft/client/Options.java
+++ b/net/minecraft/client/Options.java
@@ -752,7 +_,7 @@
         this.keyDebugLightmapTexture,
         this.keyDebugSwitchTranslucencyMode
     };
-    public KeyMapping[] keyMappings = Stream.of(
+    public final KeyMapping[] keyMappings = Stream.of(
             new KeyMapping[]{
                 this.keyAttack,
                 this.keyUse,
@@ -1993,7 +_,7 @@
         return value == 0.0 ? genericValueLabel(caption, CommonComponents.OPTION_OFF) : percentValueLabel(caption, value);
     }
 
-    public interface FieldAccess extends Options.OptionAccess {
+    private interface FieldAccess extends Options.OptionAccess {
         int process(String name, int value);
 
         boolean process(String name, boolean value);
