--- a/net/minecraft/client/SectionUpdateTracker.java
+++ b/net/minecraft/client/SectionUpdateTracker.java
@@ -66,7 +_,6 @@
         private SectionDirtyState(boolean isDirty, boolean isDirtyFromPlayer, long sectionNode) {
             this.isDirty = isDirty;
             this.isDirtyFromPlayer = isDirtyFromPlayer;
-            this.sectionNode = sectionNode;
         }
 
         public void setDirty(boolean fromPlayer) {
