--- a/net/minecraft/advancements/predicates/entity/EntityTypePredicate.java
+++ b/net/minecraft/advancements/predicates/entity/EntityTypePredicate.java
@@ -4,8 +_,8 @@
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.TagKey;
 import net.minecraft.world.entity.Entity;
@@ -14,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 public record EntityTypePredicate(HolderSet<EntityType<?>> types) implements EntitySubPredicate {
-    public static final Codec<EntityTypePredicate> CODEC = RegistryCodecs.holderSet(Registries.ENTITY_TYPE)
+    public static final Codec<EntityTypePredicate> CODEC = RegistryCodecs.homogeneousList(Registries.ENTITY_TYPE)
         .xmap(EntityTypePredicate::new, EntityTypePredicate::types);
 
     public static EntityTypePredicate of(HolderGetter<EntityType<?>> lookup, EntityType<?> type) {
