--- a/net/minecraft/advancements/predicates/entity/EntityPredicate.java
+++ b/net/minecraft/advancements/predicates/entity/EntityPredicate.java
@@ -10,13 +_,13 @@
 import java.util.Optional;
 import java.util.Map.Entry;
 import java.util.stream.Stream;
+import net.minecraft.advancements.predicates.ContextAwarePredicate;
 import net.minecraft.advancements.predicates.DataComponentMatchers;
 import net.minecraft.advancements.predicates.DistancePredicate;
 import net.minecraft.advancements.predicates.LocationPredicate;
 import net.minecraft.advancements.predicates.MobEffectsPredicate;
 import net.minecraft.advancements.predicates.NbtPredicate;
 import net.minecraft.advancements.predicates.SlotsPredicate;
-import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.component.DataComponentExactPredicate;
 import net.minecraft.core.component.predicates.DataComponentPredicate;
@@ -40,6 +_,7 @@
         BuiltInRegistries.ENTITY_SUB_PREDICATE_TYPE.byNameCodec(), c -> c
     );
     public static final Codec<EntityPredicate> CODEC = MAP_CODEC.xmap(EntityPredicate::new, p -> p.parts);
+    public static final Codec<ContextAwarePredicate> ADVANCEMENT_CODEC = Codec.withAlternative(ContextAwarePredicate.CODEC, CODEC, EntityPredicate::wrap);
     private static final Comparator<Entry<Codec<? extends EntitySubPredicate>, EntitySubPredicate>> PREDICATE_TYPE_ORDER = Entry.comparingByKey(
         Comparator.comparing(
             codec -> {
@@ -65,20 +_,21 @@
         this.combinedPart = combine(parts);
     }
 
-    public static Holder<LootItemCondition> wrap(EntityPredicate.Builder singlePredicate) {
+    public static ContextAwarePredicate wrap(EntityPredicate.Builder singlePredicate) {
         return wrap(singlePredicate.build());
     }
 
-    public static Optional<Holder<LootItemCondition>> wrap(Optional<EntityPredicate> singlePredicate) {
+    public static Optional<ContextAwarePredicate> wrap(Optional<EntityPredicate> singlePredicate) {
         return singlePredicate.map(EntityPredicate::wrap);
     }
 
-    public static List<Holder<LootItemCondition>> wrap(EntityPredicate.Builder... predicates) {
+    public static List<ContextAwarePredicate> wrap(EntityPredicate.Builder... predicates) {
         return Stream.of(predicates).map(EntityPredicate::wrap).toList();
     }
 
-    public static Holder<LootItemCondition> wrap(EntityPredicate singlePredicate) {
-        return Holder.direct(LootItemEntityPropertyCondition.hasProperties(LootContext.EntityTarget.THIS, singlePredicate).build());
+    public static ContextAwarePredicate wrap(EntityPredicate singlePredicate) {
+        LootItemCondition asCondition = LootItemEntityPropertyCondition.hasProperties(LootContext.EntityTarget.THIS, singlePredicate).build();
+        return new ContextAwarePredicate(List.of(asCondition));
     }
 
     public boolean matches(ServerPlayer player, @Nullable Entity entity) {
@@ -89,6 +_,11 @@
         return entity == null ? false : this.combinedPart.matches(entity, level, position);
     }
 
+    @Nullable
+    public <T extends EntitySubPredicate> T getPartIfExists(Codec<T> type) {
+        return (T) parts.get(type);
+    }
+
     public static LootContext createContext(ServerPlayer player, Entity entity) {
         LootParams lootParams = new LootParams.Builder(player.level())
             .withParameter(LootContextParams.THIS_ENTITY, entity)
@@ -138,6 +_,19 @@
 
         public static EntityPredicate.Builder entity() {
             return new EntityPredicate.Builder();
+        }
+
+        /**
+         * Neo: create a builder that inherits all sub-predicates from the given {@code predicate}, that match the given {@code filter}.
+         */
+        public static EntityPredicate.Builder from(EntityPredicate predicate, java.util.function.Predicate<Codec<? extends EntitySubPredicate>> filter) {
+            var builder = new Builder();
+            predicate.parts.forEach((key, value) -> {
+                if (filter.test(key)) {
+                    builder.parts.put(key, value);
+                }
+            });
+            return builder;
         }
 
         public <T extends EntitySubPredicate> EntityPredicate.Builder put(Codec<T> key, T predicate) {
