--- a/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
@@ -34,6 +34,13 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.util.Arrays;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockBrushEvent;
+// CraftBukkit end
+
 public class BrushableBlockEntity extends BlockEntity {
 
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -81,7 +88,12 @@
                     BlockState blockstate = this.getBlockState();
                     BlockState blockstate1 = (BlockState) blockstate.setValue(BlockStateProperties.DUSTED, k);
 
-                    serverlevel.setBlock(this.getBlockPos(), blockstate1, 3);
+                    // CraftBukkit start
+                    BlockBrushEvent event = CraftEventFactory.callBlockBrushEvent(serverlevel, this.getBlockPos(), blockstate1, 3, (ServerPlayer) livingentity);
+                    if (!event.isCancelled()) {
+                        event.getNewState().update(true);
+                    }
+                    // CraftBukkit end
                 }
 
                 return false;
@@ -122,10 +134,10 @@
     }
 
     private void brushingCompleted(ServerLevel serverlevel, LivingEntity livingentity, ItemStack itemstack) {
-        this.dropContent(serverlevel, livingentity, itemstack);
+        // this.dropContent(serverlevel, livingentity, itemstack); // CraftBukkit - moved down
         BlockState blockstate = this.getBlockState();
 
-        serverlevel.levelEvent(3008, this.getBlockPos(), Block.getId(blockstate));
+        // worldserver.levelEvent(3008, this.getBlockPos(), Block.getId(iblockdata)); // CraftBukkit - moved down
         Block block = this.getBlockState().getBlock();
         Block block1;
 
@@ -135,7 +147,15 @@
             block1 = Blocks.AIR;
         }
 
-        serverlevel.setBlock(this.worldPosition, block1.defaultBlockState(), 3);
+        // CraftBukkit start
+        BlockBrushEvent event = CraftEventFactory.callBlockBrushEvent(serverlevel, this.worldPosition, block1.defaultBlockState(), 3, (ServerPlayer) livingentity);
+        if (!event.isCancelled()) {
+            this.dropContent(serverlevel, livingentity, itemstack); // CraftBukkit - from above
+            serverlevel.levelEvent(3008, this.getBlockPos(), Block.getId(blockstate)); // CraftBukkit - from above
+
+            event.getNewState().update(true);
+        }
+        // CraftBukkit end
     }
 
     private void dropContent(ServerLevel serverlevel, LivingEntity livingentity, ItemStack itemstack) {
@@ -152,7 +172,12 @@
             ItemEntity itementity = new ItemEntity(serverlevel, d3, d4, d5, this.item.split(serverlevel.random.nextInt(21) + 10));
 
             itementity.setDeltaMovement(Vec3.ZERO);
-            serverlevel.addFreshEntity(itementity);
+            // CraftBukkit start
+            if (livingentity instanceof ServerPlayer serverplayer) {
+                org.bukkit.block.Block bblock = CraftBlock.at(this.level, this.worldPosition);
+                CraftEventFactory.handleBlockDropItemEvent(bblock, bblock.getState(), serverplayer, Arrays.asList(itementity));
+            }
+            // CraftBukkit end
             this.item = ItemStack.EMPTY;
         }
 
@@ -185,7 +210,7 @@
     }
 
     private boolean tryLoadLootTable(ValueInput valueinput) {
-        this.lootTable = (ResourceKey) valueinput.read("LootTable", LootTable.KEY_CODEC).orElse((Object) null);
+        this.lootTable = (ResourceKey) valueinput.read("LootTable", LootTable.KEY_CODEC).orElse(null); // CraftBukkit - decompile error
         this.lootTableSeed = valueinput.getLongOr("LootTableSeed", 0L);
         return this.lootTable != null;
     }
@@ -231,7 +256,7 @@
             this.item = ItemStack.EMPTY;
         }
 
-        this.hitDirection = (Direction) valueinput.read("hit_direction", Direction.LEGACY_ID_CODEC).orElse((Object) null);
+        this.hitDirection = (Direction) valueinput.read("hit_direction", Direction.LEGACY_ID_CODEC).orElse(null); // CraftBukkit - decompile error
     }
 
     @Override
