--- a/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkCatalystBlockEntity.java
@@ -34,10 +34,13 @@
     public SculkCatalystBlockEntity(BlockPos blockpos, BlockState blockstate) {
         super(BlockEntityType.SCULK_CATALYST, blockpos, blockstate);
         this.catalystListener = new SculkCatalystBlockEntity.CatalystListener(blockstate, new BlockPositionSource(blockpos));
+        catalystListener.level = level; // CraftBukkit
     }
 
     public static void serverTick(Level level, BlockPos blockpos, BlockState blockstate, SculkCatalystBlockEntity sculkcatalystblockentity) {
+        org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverride = sculkcatalystblockentity.getBlockPos(); // CraftBukkit - SPIGOT-7068: Add source block override, not the most elegant way but better than passing down a BlockPosition up to five methods deep.
         sculkcatalystblockentity.catalystListener.getSculkSpreader().updateCursors(level, blockpos, level.getRandom(), true);
+        org.bukkit.craftbukkit.event.CraftEventFactory.sourceBlockOverride = null; // CraftBukkit
     }
 
     @Override
@@ -63,11 +66,13 @@
         private final SculkSpreader sculkSpreader;
         private final BlockState blockState;
         private final PositionSource positionSource;
+        private Level level; // CraftBukkit
 
         public CatalystListener(BlockState blockstate, PositionSource positionsource) {
             this.blockState = blockstate;
             this.positionSource = positionsource;
             this.sculkSpreader = SculkSpreader.createLevelSpreader();
+            this.sculkSpreader.level = level; // CraftBukkit
         }
 
         @Override
