--- a/net/minecraft/world/level/block/entity/TestInstanceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TestInstanceBlockEntity.java
@@ -107,12 +107,12 @@
 
     public IChatBaseComponent getTestName() {
         return (IChatBaseComponent) this.test().map((resourcekey) -> {
-            return IChatBaseComponent.literal(resourcekey.location().toString());
+            return (IChatBaseComponent) IChatBaseComponent.literal(resourcekey.location().toString()); // CraftBukkit - decompile error
         }).orElse(TestInstanceBlockEntity.INVALID_TEST_NAME);
     }
 
     private Optional<Holder.c<GameTestInstance>> getTestHolder() {
-        Optional optional = this.test();
+        Optional<ResourceKey<GameTestInstance>> optional = this.test(); // CraftBukkit - decompile error
         IRegistryCustom iregistrycustom = this.level.registryAccess();
 
         Objects.requireNonNull(iregistrycustom);
@@ -366,7 +366,7 @@
     private void removeEntities() {
         this.level.getEntities((Entity) null, this.getStructureBounds()).stream().filter((entity) -> {
             return !(entity instanceof EntityHuman);
-        }).forEach(Entity::discard);
+        }).forEach((entity) -> entity.discard(null)); // CraftBukkit - add Bukkit remove cause
     }
 
     private void forceLoadChunks() {
