--- a/com/mojang/blaze3d/pipeline/PipelineCache.java
+++ b/com/mojang/blaze3d/pipeline/PipelineCache.java
@@ -4,10 +_,8 @@
 import com.mojang.renderpearl.api.pipeline.CompiledRenderPipeline;
 import com.mojang.renderpearl.api.pipeline.RenderPipeline;
 import com.mojang.renderpearl.api.pipeline.ShaderSource;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap;
 import java.util.Map;
-import net.minecraft.util.Util;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
@@ -23,17 +_,13 @@
         this.shaderSource = shaderSource;
     }
 
-    public void insert(RenderPipeline pipeline, CompiledRenderPipeline compiled) {
-        this.cache.put(pipeline, compiled);
-    }
-
     public @Nullable CompiledRenderPipeline get(RenderPipeline pipeline) {
         CompiledRenderPipeline cachedPipeline = this.cache.get(pipeline);
         if (cachedPipeline != null) {
             return cachedPipeline;
         }
 
-        CompiledRenderPipeline newPipeline = this.device.compilePipeline(pipeline, this.shaderSource, Util.backgroundExecutor()).join().finishCompile();
+        CompiledRenderPipeline newPipeline = this.device.compilePipeline(pipeline, this.shaderSource);
         if (newPipeline == null) {
             return null;
         }
@@ -43,13 +_,12 @@
     }
 
     public void clear() {
-        this.cache.values().forEach(UncheckedAutoCloseable::close);
+        this.cache.values().forEach(CompiledRenderPipeline::close);
         this.cache.clear();
     }
 
     @Override
     public void close() {
         this.clear();
-        this.shaderSource.close();
     }
 }
