--- a/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/net/minecraft/world/entity/AreaEffectCloud.java
@@ -107,7 +_,7 @@
         this.potionDurationScale = scale;
     }
 
-    private void updateParticle() {
+    public void updateParticle() {
         if (this.customParticle != null) {
             this.entityData.set(DATA_PARTICLE, this.customParticle);
         } else {
@@ -189,7 +_,7 @@
 
     private void serverTick(ServerLevel serverLevel) {
         if (this.duration != -1 && this.tickCount - this.waitTime >= this.duration) {
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
             boolean isWaiting = this.isWaiting();
             boolean shouldWait = this.tickCount < this.waitTime;
@@ -202,7 +_,7 @@
                 if (this.radiusPerTick != 0.0F) {
                     radius += this.radiusPerTick;
                     if (radius < 0.5F) {
-                        this.discard();
+                        this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                         return;
                     }
 
@@ -239,7 +_,7 @@
                                         if (this.radiusOnUse != 0.0F) {
                                             radius += this.radiusOnUse;
                                             if (radius < 0.5F) {
-                                                this.discard();
+                                                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                                                 return;
                                             }
 
@@ -249,7 +_,7 @@
                                         if (this.durationOnUse != 0 && this.duration != -1) {
                                             this.duration = this.duration + this.durationOnUse;
                                             if (this.duration <= 0) {
-                                                this.discard();
+                                                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                                                 return;
                                             }
                                         }
