--- a/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/net/minecraft/world/entity/AreaEffectCloud.java
@@ -23,6 +_,7 @@
 import net.minecraft.world.level.material.PushReaction;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
 import org.jspecify.annotations.Nullable;
 
 public class AreaEffectCloud extends Entity implements TraceableEntity {
@@ -107,7 +_,7 @@
         this.potionDurationScale = scale;
     }
 
-    private void updateParticle() {
+    public void updateParticle() {
         if (this.customParticle != null) {
             this.entityData.set(DATA_PARTICLE, this.customParticle);
         } else {
@@ -189,7 +_,7 @@
 
     private void serverTick(ServerLevel serverLevel) {
         if (this.duration != -1 && this.tickCount - this.waitTime >= this.duration) {
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
             boolean isWaiting = this.isWaiting();
             boolean shouldWait = this.tickCount < this.waitTime;
@@ -202,7 +_,7 @@
                 if (this.radiusPerTick != 0.0F) {
                     radius += this.radiusPerTick;
                     if (radius < 0.5F) {
-                        this.discard();
+                        this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                         return;
                     }
 
@@ -218,39 +_,57 @@
                         this.potionContents.forEachEffect(allEffects::add, this.potionDurationScale);
                         List<LivingEntity> entities = this.level().getEntitiesOfClass(LivingEntity.class, this.getBoundingBox());
                         if (!entities.isEmpty()) {
-                            for (LivingEntity entity : entities) {
-                                if (!this.victims.containsKey(entity) && entity.isAffectedByPotions() && !allEffects.stream().noneMatch(entity::canBeAffected)) {
-                                    double xd = entity.getX() - this.getX();
-                                    double zd = entity.getZ() - this.getZ();
-                                    double dist = xd * xd + zd * zd;
-                                    if (dist <= radius * radius) {
-                                        this.victims.put(entity, this.tickCount + this.reapplicationDelay);
-
-                                        for (MobEffectInstance effect : allEffects) {
-                                            if (effect.getEffect().value().isInstantenous()) {
-                                                effect.getEffect()
-                                                    .value()
-                                                    .applyInstantenousEffect(serverLevel, this, this.getOwner(), entity, effect.getAmplifier(), 0.5);
-                                            } else {
-                                                entity.addEffect(new MobEffectInstance(effect), this);
-                                            }
-                                        }
-
-                                        if (this.radiusOnUse != 0.0F) {
-                                            radius += this.radiusOnUse;
-                                            if (radius < 0.5F) {
-                                                this.discard();
-                                                return;
-                                            }
-
-                                            this.setRadius(radius);
-                                        }
-
-                                        if (this.durationOnUse != 0 && this.duration != -1) {
-                                            this.duration = this.duration + this.durationOnUse;
-                                            if (this.duration <= 0) {
-                                                this.discard();
-                                                return;
+                            List<org.bukkit.entity.LivingEntity> bukkitEntities = new java.util.ArrayList<>(); // CraftBukkit
+                            for (LivingEntity livingentity : entities) {
+                                if (!this.victims.containsKey(livingentity) && livingentity.isAffectedByPotions()) {
+                                    java.util.stream.Stream<MobEffectInstance> stream = allEffects.stream(); // CraftBukkit - decompile error
+
+                                    java.util.Objects.requireNonNull(livingentity);
+                                    if (!stream.noneMatch(livingentity::canBeAffected)) {
+                                        double d0 = livingentity.getX() - this.getX();
+                                        double d1 = livingentity.getZ() - this.getZ();
+                                        double d2 = d0 * d0 + d1 * d1;
+
+                                        if (d2 <= (double) (radius * radius)) {
+                                            // CraftBukkit start
+                                            bukkitEntities.add((org.bukkit.entity.LivingEntity) livingentity.getBukkitEntity());
+                                        }
+                                    }
+                                }
+                            }
+                            {
+                                org.bukkit.event.entity.AreaEffectCloudApplyEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callAreaEffectCloudApplyEvent(this, bukkitEntities);
+                                if (!event.isCancelled()) {
+                                    for (org.bukkit.entity.LivingEntity entity : event.getAffectedEntities()) {
+                                        if (entity instanceof CraftLivingEntity) {
+                                            LivingEntity livingentity = ((CraftLivingEntity) entity).getHandle();
+                                            // CraftBukkit end
+                                            this.victims.put(livingentity, this.tickCount + this.reapplicationDelay);
+
+                                            for (MobEffectInstance mobeffectinstance : allEffects) {
+                                                if (mobeffectinstance.getEffect().value().isInstantenous()) {
+                                                    mobeffectinstance.getEffect().value().applyInstantenousEffect(serverLevel, this, this.getOwner(), livingentity, mobeffectinstance.getAmplifier(), 0.5);
+                                                } else {
+                                                    livingentity.addEffect(new MobEffectInstance(mobeffectinstance), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.AREA_EFFECT_CLOUD); // CraftBukkit
+                                                }
+                                            }
+
+                                            if (this.radiusOnUse != 0.0F) {
+                                                radius += this.radiusOnUse;
+                                                if (radius < 0.5F) {
+                                                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                                                    return;
+                                                }
+
+                                                this.setRadius(radius);
+                                            }
+
+                                            if (this.durationOnUse != 0 && this.duration != -1) {
+                                                this.duration = this.duration + this.durationOnUse;
+                                                if (this.duration <= 0) {
+                                                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                                                    return;
+                                                }
                                             }
                                         }
                                     }
