--- a/net/minecraft/world/entity/AreaEffectCloud.java
+++ b/net/minecraft/world/entity/AreaEffectCloud.java
@@ -38,14 +_,14 @@
     private static final float DEFAULT_RADIUS = 3.0F;
     public static final float DEFAULT_WIDTH = 6.0F;
     public static final float HEIGHT = 0.5F;
-    public PotionContents potionContents = PotionContents.EMPTY;
+    private PotionContents potionContents = PotionContents.EMPTY;
     private final Map<Entity, Integer> victims = Maps.newHashMap();
     private int duration = 600;
-    public int waitTime = 20;
-    public int reapplicationDelay = 20;
-    public int durationOnUse;
-    public float radiusOnUse;
-    public float radiusPerTick;
+    private int waitTime = 20;
+    private int reapplicationDelay = 20;
+    private int durationOnUse;
+    private float radiusOnUse;
+    private float radiusPerTick;
     @Nullable
     private LivingEntity owner;
     @Nullable
@@ -92,7 +_,7 @@
         this.updateColor();
     }
 
-    public void updateColor() {
+    private void updateColor() {
         ParticleOptions particleoptions = this.entityData.get(DATA_PARTICLE);
         if (particleoptions instanceof ColorParticleOption colorparticleoption) {
             int i = this.potionContents.equals(PotionContents.EMPTY) ? 0 : this.potionContents.getColor();
