--- a/net/minecraft/world/entity/ConversionTracker.java
+++ b/net/minecraft/world/entity/ConversionTracker.java
@@ -23,8 +_,8 @@
     private final String conversionTimeTag;
     private final int totalConversionTime;
     private final @Nullable BiConsumer<M, ServerLevel> postConversionStep;
-    public int afflictionTime = 0;
-    public int conversionTime;
+    private int afflictionTime = 0;
+    private int conversionTime;
 
     public ConversionTracker(
         M mob,
@@ -75,7 +_,7 @@
         return this.mob.getEntityData().get(this.conversionId);
     }
 
-    public void setConverting(boolean isConverting) {
+    private void setConverting(boolean isConverting) {
         this.mob.getEntityData().set(this.conversionId, isConverting);
     }
 
