--- a/net/minecraft/world/entity/Display.java
+++ b/net/minecraft/world/entity/Display.java
@@ -895,7 +_,7 @@
             b0 = loadFlag(b0, p_270714_, "default_background", (byte)4);
             Optional<Display.TextDisplay.Align> optional = Display.TextDisplay.Align.CODEC
                 .decode(NbtOps.INSTANCE, p_270714_.get("alignment"))
-                .resultOrPartial(Util.prefix("Display entity", Display.LOGGER::error))
+                .result() // Paper - Hide text display error on spawn
                 .map(Pair::getFirst);
             if (optional.isPresent()) {
                 b0 = switch ((Display.TextDisplay.Align)optional.get()) {
