--- a/net/minecraft/world/entity/EntityFluidInteraction.java
+++ b/net/minecraft/world/entity/EntityFluidInteraction.java
@@ -6,6 +_,7 @@
 import java.util.Map.Entry;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
+import net.minecraft.tags.FluidTags;
 import net.minecraft.tags.TagKey;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.player.Player;
@@ -79,6 +_,11 @@
                                             tracker.accumulateCurrent(flow);
                                         }
                                     }
+                                    // CraftBukkit start - store last lava contact location
+                                    if (fluidType.is(FluidTags.LAVA)) {
+                                        entity.lastLavaContact = mutablePos.immutable();
+                                    }
+                                    // CraftBukkit end
                                 }
                             }
                         }
