--- a/net/minecraft/world/entity/EntityFluidInteraction.java
+++ b/net/minecraft/world/entity/EntityFluidInteraction.java
@@ -1,15 +_,12 @@
 package net.minecraft.world.entity;
 
 import it.unimi.dsi.fastutil.objects.Reference2ObjectArrayMap;
-import it.unimi.dsi.fastutil.objects.Reference2ObjectMap;
-import it.unimi.dsi.fastutil.objects.Reference2ObjectMaps;
-import it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry;
-import java.util.ArrayList;
-import java.util.List;
+import java.util.Map;
 import java.util.Set;
+import java.util.Map.Entry;
 import net.minecraft.core.BlockPos;
-import net.minecraft.core.Holder;
 import net.minecraft.core.SectionPos;
+import net.minecraft.tags.FluidTags;
 import net.minecraft.tags.TagKey;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.player.Player;
@@ -25,82 +_,76 @@
 import org.jspecify.annotations.Nullable;
 
 public class EntityFluidInteraction {
-    private final List<EntityFluidInteraction.Tracker> fluidTrackers = new ArrayList<>();
-    private final Reference2ObjectMap<TagKey<Fluid>, EntityFluidInteraction.CurrentAccumulator> currentAccumulators = new Reference2ObjectArrayMap<>();
+    private final Map<TagKey<Fluid>, EntityFluidInteraction.Tracker> trackerByFluid = new Reference2ObjectArrayMap<>();
 
-    public EntityFluidInteraction(Set<TagKey<Fluid>> fluidsWithCurrent) {
-        for (TagKey<Fluid> fluid : fluidsWithCurrent) {
-            this.currentAccumulators.put(fluid, new EntityFluidInteraction.CurrentAccumulator());
+    public EntityFluidInteraction(Set<TagKey<Fluid>> fluids) {
+        for (TagKey<Fluid> fluid : fluids) {
+            this.trackerByFluid.put(fluid, new EntityFluidInteraction.Tracker());
         }
     }
 
-    public boolean update(Entity entity, boolean ignoreCurrent) {
-        this.fluidTrackers.removeIf(EntityFluidInteraction.Tracker::reset);
-        this.currentAccumulators.values().forEach(EntityFluidInteraction.CurrentAccumulator::reset);
+    public void update(Entity entity, boolean ignoreCurrent) {
+        this.trackerByFluid.values().forEach(EntityFluidInteraction.Tracker::reset);
         AABB box = entity.getFluidInteractionBox();
-        if (box == null) {
-            return false;
-        }
-
-        int x0 = Mth.floor(box.minX);
-        int y0 = Mth.floor(box.minY);
-        int z0 = Mth.floor(box.minZ);
-        int x1 = Mth.ceil(box.maxX) - 1;
-        int y1 = Mth.ceil(box.maxY) - 1;
-        int z1 = Mth.ceil(box.maxZ) - 1;
-        if (!hasFluidAndLoaded(entity.level(), x0 - 1, y0, z0 - 1, x1 + 1, y1, z1 + 1)) {
-            return false;
-        }
-
-        double entityY = entity.getBoundingBox().minY;
-        int eyeBlockX = entity.getBlockX();
-        double eyeY = entity.getEyeY();
-        int eyeBlockZ = entity.getBlockZ();
-        Holder<Fluid> lastFluidType = null;
-        EntityFluidInteraction.Tracker tracker = null;
-        EntityFluidInteraction.CurrentAccumulator current = null;
-        BlockGetter level = entity.level();
-        BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
-
-        for (int x = x0; x <= x1; x++) {
-            for (int y = y0; y <= y1; y++) {
-                for (int z = z0; z <= z1; z++) {
-                    mutablePos.set(x, y, z);
-                    FluidState fluidState = level.getFluidState(mutablePos);
-                    if (!fluidState.isEmpty()) {
-                        double fluidBottom = mutablePos.getY();
-                        double fluidTop = fluidBottom + fluidState.getHeight(level, mutablePos);
-                        if (!(fluidTop < box.minY)) {
-                            Holder<Fluid> fluidType = fluidState.typeHolder();
-                            if (fluidType != lastFluidType) {
-                                lastFluidType = fluidType;
-                                tracker = this.getOrCreateTrackerFor(fluidType);
-                                if (!ignoreCurrent) {
-                                    current = this.getCurrentAccumulatorFor(fluidType);
-                                }
-                            }
-
-                            if (x == eyeBlockX && z == eyeBlockZ && eyeY >= fluidBottom && eyeY <= fluidTop) {
-                                tracker.eyesInside = true;
-                            }
-
-                            tracker.height = Math.max(fluidTop - entityY, tracker.height);
-                            if (current != null) {
-                                Vec3 flow = fluidState.getFlow(level, mutablePos);
-                                current.height = Math.max(tracker.height, current.height);
-                                if (current.height < 0.4) {
-                                    flow = flow.scale(current.height);
-                                }
-
-                                current.accumulate(flow);
+        if (box != null) {
+            int x0 = Mth.floor(box.minX);
+            int y0 = Mth.floor(box.minY);
+            int z0 = Mth.floor(box.minZ);
+            int x1 = Mth.ceil(box.maxX) - 1;
+            int y1 = Mth.ceil(box.maxY) - 1;
+            int z1 = Mth.ceil(box.maxZ) - 1;
+            if (hasFluidAndLoaded(entity.level(), x0 - 1, y0, z0 - 1, x1 + 1, y1, z1 + 1)) {
+                double entityY = entity.getBoundingBox().minY;
+                int eyeBlockX = entity.getBlockX();
+                double eyeY = entity.getEyeY();
+                int eyeBlockZ = entity.getBlockZ();
+                Fluid lastFluidType = null;
+                EntityFluidInteraction.Tracker tracker = null;
+                BlockGetter level = entity.level();
+                BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
+
+                for (int x = x0; x <= x1; x++) {
+                    for (int y = y0; y <= y1; y++) {
+                        for (int z = z0; z <= z1; z++) {
+                            mutablePos.set(x, y, z);
+                            FluidState fluidState = level.getFluidState(mutablePos);
+                            if (!fluidState.isEmpty()) {
+                                double fluidBottom = mutablePos.getY();
+                                double fluidTop = fluidBottom + fluidState.getHeight(level, mutablePos);
+                                if (!(fluidTop < box.minY)) {
+                                    Fluid fluidType = fluidState.getType();
+                                    if (fluidType != lastFluidType) {
+                                        lastFluidType = fluidType;
+                                        tracker = this.getTrackerFor(fluidType);
+                                    }
+
+                                    if (tracker != null) {
+                                        if (x == eyeBlockX && z == eyeBlockZ && eyeY >= fluidBottom && eyeY <= fluidTop) {
+                                            tracker.eyesInside = true;
+                                        }
+
+                                        tracker.height = Math.max(fluidTop - entityY, tracker.height);
+                                        if (!ignoreCurrent) {
+                                            Vec3 flow = fluidState.getFlow(level, mutablePos);
+                                            if (tracker.height < 0.4) {
+                                                flow = flow.scale(tracker.height);
+                                            }
+
+                                            tracker.accumulateCurrent(flow);
+                                        }
+                                    }
+                                    // CraftBukkit start - store last lava contact location
+                                    if (fluidType.is(FluidTags.LAVA)) {
+                                        entity.lastLavaContact = mutablePos.immutable();
+                                    }
+                                    // CraftBukkit end
+                                }
                             }
                         }
                     }
                 }
             }
         }
-
-        return lastFluidType != null;
     }
 
     private static boolean hasFluidAndLoaded(Level level, int x0, int y0, int z0, int x1, int y1, int z1) {
@@ -133,21 +_,10 @@
         return hasFluid;
     }
 
-    private EntityFluidInteraction.Tracker getOrCreateTrackerFor(Holder<Fluid> fluidType) {
-        for (EntityFluidInteraction.Tracker tracker : this.fluidTrackers) {
-            if (tracker.fluidType.equals(fluidType)) {
-                return tracker;
-            }
-        }
-
-        EntityFluidInteraction.Tracker tracker = new EntityFluidInteraction.Tracker(fluidType);
-        this.fluidTrackers.add(tracker);
-        return tracker;
-    }
-
-    private EntityFluidInteraction.@Nullable CurrentAccumulator getCurrentAccumulatorFor(Holder<Fluid> fluid) {
-        for (Entry<TagKey<Fluid>, EntityFluidInteraction.CurrentAccumulator> entry : Reference2ObjectMaps.fastIterable(this.currentAccumulators)) {
-            if (fluid.is(entry.getKey())) {
+    private EntityFluidInteraction.@Nullable Tracker getTrackerFor(Fluid fluid) {
+        for (Entry<TagKey<Fluid>, EntityFluidInteraction.Tracker> entry : this.trackerByFluid.entrySet()) {
+            TagKey<Fluid> tag = entry.getKey();
+            if (fluid.is(tag)) {
                 return entry.getValue();
             }
         }
@@ -156,22 +_,15 @@
     }
 
     public void applyCurrentTo(TagKey<Fluid> fluid, Entity entity, double scale) {
-        EntityFluidInteraction.CurrentAccumulator current = this.currentAccumulators.get(fluid);
-        if (current != null) {
-            current.applyTo(entity, scale);
+        EntityFluidInteraction.Tracker tracker = this.trackerByFluid.get(fluid);
+        if (tracker != null) {
+            tracker.applyCurrentTo(entity, scale);
         }
     }
 
     public double getFluidHeight(TagKey<Fluid> fluid) {
-        double height = 0.0;
-
-        for (EntityFluidInteraction.Tracker tracker : this.fluidTrackers) {
-            if (tracker.height > height && tracker.fluidType.is(fluid)) {
-                height = tracker.height;
-            }
-        }
-
-        return height;
+        EntityFluidInteraction.Tracker tracker = this.trackerByFluid.get(fluid);
+        return tracker != null ? tracker.height : 0.0;
     }
 
     public boolean isInFluid(TagKey<Fluid> fluid) {
@@ -179,32 +_,29 @@
     }
 
     public boolean isEyeInFluid(TagKey<Fluid> fluid) {
-        for (EntityFluidInteraction.Tracker tracker : this.fluidTrackers) {
-            if (tracker.eyesInside && tracker.fluidType.is(fluid)) {
-                return true;
-            }
-        }
-
-        return false;
+        EntityFluidInteraction.Tracker tracker = this.trackerByFluid.get(fluid);
+        return tracker != null && tracker.eyesInside;
     }
 
-    private static class CurrentAccumulator {
+    private static class Tracker {
         private double height;
+        private boolean eyesInside;
         private Vec3 accumulatedCurrent = Vec3.ZERO;
         private int currentCount;
 
         public void reset() {
             this.height = 0.0;
+            this.eyesInside = false;
             this.accumulatedCurrent = Vec3.ZERO;
             this.currentCount = 0;
         }
 
-        public void accumulate(Vec3 flow) {
+        public void accumulateCurrent(Vec3 flow) {
             this.accumulatedCurrent = this.accumulatedCurrent.add(flow);
             this.currentCount++;
         }
 
-        public void applyTo(Entity entity, double scale) {
+        public void applyCurrentTo(Entity entity, double scale) {
             if (this.currentCount != 0 && !(this.accumulatedCurrent.lengthSqr() < 1.0E-5F)) {
                 Vec3 impulse;
                 if (!(entity instanceof Player)) {
@@ -222,26 +_,6 @@
 
                 entity.addDeltaMovement(impulse);
             }
-        }
-    }
-
-    private static class Tracker {
-        private final Holder<Fluid> fluidType;
-        private double height;
-        private boolean eyesInside;
-
-        public Tracker(Holder<Fluid> fluidType) {
-            this.fluidType = fluidType;
-        }
-
-        public boolean reset() {
-            if (this.height == 0.0 && !this.eyesInside) {
-                return true;
-            }
-
-            this.height = 0.0;
-            this.eyesInside = false;
-            return false;
         }
     }
 }
