--- a/net/minecraft/world/entity/EntitySelector.java
+++ b/net/minecraft/world/entity/EntitySelector.java
@@ -17,10 +_,17 @@
             || !p_20436_.isSpectator() && !((Player)p_20436_).isCreative();
     public static final Predicate<Entity> NO_SPECTATORS = p_20434_ -> !p_20434_.isSpectator();
     public static final Predicate<Entity> CAN_BE_COLLIDED_WITH = NO_SPECTATORS.and(Entity::canBeCollidedWith);
-
+    public static Predicate<Player> IS_INSOMNIAC = (player) -> net.minecraft.util.Mth.clamp(((net.minecraft.server.level.ServerPlayer) player).getStats().getValue(net.minecraft.stats.Stats.CUSTOM.get(net.minecraft.stats.Stats.TIME_SINCE_REST)), 1, Integer.MAX_VALUE) >= 72000; // Paper - Add phantom creative and insomniac controls
+    public static Predicate<Player> notAfk = (player) -> !player.isAfk(); // Purpur
     private EntitySelector() {
     }
 
+    // Paper start - Affects Spawning API
+    public static final Predicate<Entity> PLAYER_AFFECTS_SPAWNING = (entity) -> {
+        return !entity.isSpectator() && entity.isAlive() && entity instanceof Player player && player.affectsSpawning;
+    };
+    // Paper end - Affects Spawning API
+
     public static Predicate<Entity> withinDistance(double p_20411_, double p_20412_, double p_20413_, double p_20414_) {
         double d0 = p_20414_ * p_20414_;
         return p_20420_ -> p_20420_ != null && p_20420_.distanceToSqr(p_20411_, p_20412_, p_20413_) <= d0;
@@ -33,9 +_,9 @@
             ? Predicates.alwaysFalse()
             : NO_SPECTATORS.and(
                 p_20430_ -> {
-                    if (!p_20430_.isPushable()) {
+                    if (!p_20430_.canCollideWithBukkit(p_20422_) || !p_20422_.canCollideWithBukkit(p_20430_)) { // CraftBukkit - collidable API
                         return false;
-                    } else if (!p_20422_.level().isClientSide || p_20430_ instanceof Player && ((Player)p_20430_).isLocalPlayer()) {
+                    } else if (!(p_20430_ instanceof Player) || !(p_20422_ instanceof Player) || io.papermc.paper.configuration.GlobalConfiguration.get().collisions.enablePlayerCollisions) {
                         Team team1 = p_20430_.getTeam();
                         Team.CollisionRule team$collisionrule1 = team1 == null ? Team.CollisionRule.ALWAYS : team1.getCollisionRule();
                         if (team$collisionrule1 == Team.CollisionRule.NEVER) {
