--- a/net/minecraft/world/entity/EntityTypes.java
+++ b/net/minecraft/world/entity/EntityTypes.java
@@ -67,7 +_,7 @@
 import net.minecraft.world.entity.monster.Blaze;
 import net.minecraft.world.entity.monster.Creeper;
 import net.minecraft.world.entity.monster.ElderGuardian;
-import net.minecraft.world.entity.monster.Enderman;
+import net.minecraft.world.entity.monster.EnderMan;
 import net.minecraft.world.entity.monster.Endermite;
 import net.minecraft.world.entity.monster.Ghast;
 import net.minecraft.world.entity.monster.Giant;
@@ -179,7 +_,7 @@
             .fireImmune()
             .sized(6.0F, 0.5F)
             .clientTrackingRange(10)
-            .noUpdateInterval()
+            .updateInterval(Integer.MAX_VALUE)
     );
     public static final EntityType<Armadillo> ARMADILLO = register(
         EntityTypeIds.ARMADILLO, EntityType.Builder.of(Armadillo::new, MobCategory.CREATURE).sized(0.7F, 0.65F).eyeHeight(0.26F).clientTrackingRange(10)
@@ -387,9 +_,9 @@
             .clientTrackingRange(10)
             .notInPeaceful()
     );
-    public static final EntityType<Enderman> ENDERMAN = register(
+    public static final EntityType<EnderMan> ENDERMAN = register(
         EntityTypeIds.ENDERMAN,
-        EntityType.Builder.of(Enderman::new, MobCategory.MONSTER)
+        EntityType.Builder.of(EnderMan::new, MobCategory.MONSTER)
             .sized(0.6F, 2.9F)
             .eyeHeight(2.55F)
             .passengerAttachments(2.80625F)
@@ -424,7 +_,7 @@
             .fireImmune()
             .sized(2.0F, 2.0F)
             .clientTrackingRange(16)
-            .noUpdateInterval()
+            .updateInterval(Integer.MAX_VALUE)
             .dontTrackDeltas()
     );
     public static final EntityType<Evoker> EVOKER = register(
@@ -534,7 +_,7 @@
             .sized(0.5F, 0.5F)
             .eyeHeight(0.0F)
             .clientTrackingRange(10)
-            .noUpdateInterval()
+            .updateInterval(Integer.MAX_VALUE)
             .dontTrackDeltas()
     );
     public static final EntityType<GlowSquid> GLOW_SQUID = register(
@@ -610,7 +_,7 @@
             .sized(0.5F, 0.5F)
             .eyeHeight(0.0F)
             .clientTrackingRange(10)
-            .noUpdateInterval()
+            .updateInterval(Integer.MAX_VALUE)
             .dontTrackDeltas()
     );
     public static final EntityType<Boat> JUNGLE_BOAT = register(
@@ -637,12 +_,17 @@
             .sized(0.375F, 0.5F)
             .eyeHeight(0.0625F)
             .clientTrackingRange(10)
-            .noUpdateInterval()
+            .updateInterval(Integer.MAX_VALUE)
             .dontTrackDeltas()
     );
     public static final EntityType<LightningBolt> LIGHTNING_BOLT = register(
         EntityTypeIds.LIGHTNING_BOLT,
-        EntityType.Builder.of(LightningBolt::new, MobCategory.MISC).noLootTable().noSave().sized(0.0F, 0.0F).clientTrackingRange(16).noUpdateInterval()
+        EntityType.Builder.of(LightningBolt::new, MobCategory.MISC)
+            .noLootTable()
+            .noSave()
+            .sized(0.0F, 0.0F)
+            .clientTrackingRange(16)
+            .updateInterval(Integer.MAX_VALUE)
     );
     public static final EntityType<Llama> LLAMA = register(
         EntityTypeIds.LLAMA,
@@ -748,7 +_,7 @@
             .noLootTable()
             .sized(0.5F, 0.5F)
             .clientTrackingRange(10)
-            .noUpdateInterval()
+            .updateInterval(Integer.MAX_VALUE)
             .dontTrackDeltas()
     );
     public static final EntityType<Boat> PALE_OAK_BOAT = register(
