--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -225,7 +_,7 @@
             .fireImmune()
             .sized(6.0F, 0.5F)
             .clientTrackingRange(10)
-            .updateInterval(Integer.MAX_VALUE)
+            .updateInterval(10) // CraftBukkit - SPIGOT-3729: track area effect clouds
     );
     public static final EntityType<Armadillo> ARMADILLO = register(
         "armadillo", EntityType.Builder.of(Armadillo::new, MobCategory.CREATURE).sized(0.7F, 0.65F).eyeHeight(0.26F).clientTrackingRange(10)
@@ -1225,6 +_,11 @@
     private final FeatureFlagSet requiredFeatures;
     private final boolean allowedInPeaceful;
 
+    private final java.util.function.Predicate<EntityType<?>> trackDeltasSupplier;
+    private final java.util.function.ToIntFunction<EntityType<?>> trackingRangeSupplier;
+    private final java.util.function.ToIntFunction<EntityType<?>> updateIntervalSupplier;
+    private final boolean onlyOpCanSetNbt;
+
     private static <T extends Entity> EntityType<T> register(ResourceKey<EntityType<?>> id, EntityType.Builder<T> builder) {
         return Registry.register(BuiltInRegistries.ENTITY_TYPE, id, builder.build(id));
     }
@@ -1233,7 +_,7 @@
         return ResourceKey.create(Registries.ENTITY_TYPE, Identifier.withDefaultNamespace(vanillaId));
     }
 
-    private static <T extends Entity> EntityType<T> register(String vanillaId, EntityType.Builder<T> builder) {
+    private static <T extends Entity> EntityType<T> register(String vanillaId, EntityType.Builder builder) { // CraftBukkit - decompile error
         return register(vanillaEntityId(vanillaId), builder);
     }
 
@@ -1262,6 +_,30 @@
         FeatureFlagSet requiredFeatures,
         boolean allowedInPeaceful
     ) {
+        this(factory, category, serialize, summon, fireImmune, canSpawnFarFromPlayer, immuneTo, dimensions, spawnDimensionsScale, clientTrackingRange, updateInterval, descriptionId, lootTable, requiredFeatures, allowedInPeaceful, EntityType::defaultTrackDeltasSupplier, EntityType::defaultTrackingRangeSupplier, EntityType::defaultUpdateIntervalSupplier, false);
+    }
+
+    public EntityType(
+            EntityType.EntityFactory<T> factory,
+            MobCategory category,
+            boolean serialize,
+            boolean summon,
+            boolean fireImmune,
+            boolean canSpawnFarFromPlayer,
+            ImmutableSet<Block> immuneTo,
+            EntityDimensions dimensions,
+            float spawnDimensionsScale,
+            int clientTrackingRange,
+            int updateInterval,
+            String descriptionId,
+            Optional<ResourceKey<LootTable>> lootTable,
+            FeatureFlagSet requiredFeatures,
+            boolean allowedInPeaceful,
+            final java.util.function.Predicate<EntityType<?>> trackDeltasSupplier,
+            final java.util.function.ToIntFunction<EntityType<?>> trackingRangeSupplier,
+            final java.util.function.ToIntFunction<EntityType<?>> updateIntervalSupplier,
+            boolean onlyOpCanSetNbt
+    ) {
         this.factory = factory;
         this.category = category;
         this.canSpawnFarFromPlayer = canSpawnFarFromPlayer;
@@ -1277,25 +_,28 @@
         this.lootTable = lootTable;
         this.requiredFeatures = requiredFeatures;
         this.allowedInPeaceful = allowedInPeaceful;
-    }
-
-    public @Nullable T spawn(
-        ServerLevel level,
-        @Nullable ItemStack itemStack,
-        @Nullable LivingEntity user,
-        BlockPos spawnPos,
-        EntitySpawnReason spawnReason,
-        boolean tryMoveDown,
-        boolean movedUp
-    ) {
-        Consumer<T> postSpawnConfig;
+        this.trackDeltasSupplier = trackDeltasSupplier;
+        this.trackingRangeSupplier = trackingRangeSupplier;
+        this.updateIntervalSupplier = updateIntervalSupplier;
+        this.onlyOpCanSetNbt = onlyOpCanSetNbt;
+    }
+
+    public @Nullable T spawn(ServerLevel level, @Nullable ItemStack itemStack, @Nullable LivingEntity user, BlockPos spawnPos, EntitySpawnReason spawnReason, boolean tryMoveDown, boolean movedUp) {
+        // CraftBukkit start
+        return this.spawn(level, itemStack, user, spawnPos, spawnReason, tryMoveDown, movedUp, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG);
+    }
+
+    public @Nullable T spawn(ServerLevel level, @Nullable ItemStack itemStack, @Nullable LivingEntity user, BlockPos spawnPos, EntitySpawnReason spawnReason, boolean tryMoveDown, boolean movedUp, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason bukkitSpawnReason) {
+        // CraftBukkit end
+        Consumer<T> consumer;
+
         if (itemStack != null) {
-            postSpawnConfig = createDefaultStackConfig(level, itemStack, user);
+            consumer = createDefaultStackConfig(level, itemStack, user);
         } else {
-            postSpawnConfig = entity -> {};
+            consumer = entity -> {};
         }
 
-        return this.spawn(level, postSpawnConfig, spawnPos, spawnReason, tryMoveDown, movedUp);
+        return (T) this.spawn(level, consumer, spawnPos, spawnReason, tryMoveDown, movedUp, bukkitSpawnReason); // CraftBukkit
     }
 
     public static <T extends Entity> Consumer<T> createDefaultStackConfig(Level level, ItemStack itemStack, @Nullable LivingEntity user) {
@@ -1315,20 +_,38 @@
     public static <T extends Entity> Consumer<T> appendCustomEntityStackConfig(
         Consumer<T> initialConfig, Level level, ItemStack itemStack, @Nullable LivingEntity user
     ) {
-        TypedEntityData<EntityType<?>> entityData = itemStack.get(DataComponents.ENTITY_DATA);
-        return entityData != null ? initialConfig.andThen(entity -> updateCustomEntityTag(level, user, entity, entityData)) : initialConfig;
+        TypedEntityData<EntityType<?>> typedentitydata = (TypedEntityData) itemStack.get(DataComponents.ENTITY_DATA);
+
+        return typedentitydata != null ? initialConfig.andThen((entity) -> {
+            try { updateCustomEntityTag(level, user, entity, typedentitydata); } catch (Throwable t) { LOGGER.warn("Error loading spawn egg NBT", t); } // CraftBukkit - SPIGOT-5665
+        }) : initialConfig;
     }
 
     public @Nullable T spawn(ServerLevel level, BlockPos spawnPos, EntitySpawnReason spawnReason) {
-        return this.spawn(level, null, spawnPos, spawnReason, false, false);
-    }
-
-    public @Nullable T spawn(
-        ServerLevel level, @Nullable Consumer<T> postSpawnConfig, BlockPos spawnPos, EntitySpawnReason spawnReason, boolean tryMoveDown, boolean movedUp
-    ) {
+        // CraftBukkit start
+        return (T) this.spawn(level, spawnPos, spawnReason, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public @Nullable T spawn(ServerLevel level, BlockPos spawnPos, EntitySpawnReason spawnReason, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason bukkitSpawnReason) {
+        return (T) this.spawn(level, (Consumer) null, spawnPos, spawnReason, false, false, bukkitSpawnReason); // CraftBukkit
+        // CraftBukkit end
+    }
+
+    public @Nullable T spawn(ServerLevel level, @Nullable Consumer<T> postSpawnConfig, BlockPos spawnPos, EntitySpawnReason spawnReason, boolean tryMoveDown, boolean movedUp) {
+        // CraftBukkit start
+        return this.spawn(level, postSpawnConfig, spawnPos, spawnReason, tryMoveDown, movedUp, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public @Nullable T spawn(ServerLevel level, @Nullable Consumer<T> postSpawnConfig, BlockPos spawnPos, EntitySpawnReason spawnReason, boolean tryMoveDown, boolean movedUp, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason bukkitSpawnReason) {
+        // CraftBukkit end
         T entity = this.create(level, postSpawnConfig, spawnPos, spawnReason, tryMoveDown, movedUp);
         if (entity != null) {
-            level.addFreshEntityWithPassengers(entity);
+            // CraftBukkit start
+            level.addFreshEntityWithPassengers(entity, bukkitSpawnReason);
+            if (entity.isRemoved()) {
+                return null; // Don't return an entity when CreatureSpawnEvent is canceled
+            }
+            // CraftBukkit end
             if (entity instanceof Mob mob) {
                 mob.playAmbientSound();
             }
@@ -1357,6 +_,13 @@
                 mob.yHeadRot = mob.getYRot();
                 mob.yBodyRot = mob.getYRot();
                 mob.finalizeSpawn(level, level.getCurrentDifficultyAt(mob.blockPosition()), spawnReason, null);
+                if (mob.isSpawnCancelled()) {
+                    // Neo: Discard mob, spawn was cancelled
+                    mob.discard();
+                    // return null, mob was killed, context should be lost
+                    // fixes llamas for wandering trader spawning if wandering trader was cancelled
+                    return null;
+                }
             }
 
             if (postSpawnConfig != null) {
@@ -1560,14 +_,23 @@
     }
 
     public int clientTrackingRange() {
+        return trackingRangeSupplier.applyAsInt(this);
+    }
+    private int defaultTrackingRangeSupplier() {
         return this.clientTrackingRange;
     }
 
     public int updateInterval() {
+        return updateIntervalSupplier.applyAsInt(this);
+    }
+    private int defaultUpdateIntervalSupplier() {
         return this.updateInterval;
     }
 
     public boolean trackDeltas() {
+        return this.trackDeltasSupplier.test(this);
+    }
+    private boolean defaultTrackDeltasSupplier() {
         return this != PLAYER
             && this != LLAMA_SPIT
             && this != WITHER
@@ -1615,9 +_,12 @@
     }
 
     public boolean onlyOpCanSetNbt() {
+        if (onlyOpCanSetNbt) return true;
         return OP_ONLY_CUSTOM_DATA.contains(this);
     }
 
+    public Stream<net.minecraft.tags.TagKey<EntityType<?>>> getTags() {return this.builtInRegistryHolder().tags();}
+
     public static class Builder<T extends Entity> {
         private final EntityType.EntityFactory<T> factory;
         private final MobCategory category;
@@ -1637,6 +_,10 @@
         );
         private final DependantName<EntityType<?>, String> descriptionId = id -> Util.makeDescriptionId("entity", id.identifier());
         private boolean allowedInPeaceful = true;
+        private java.util.function.Predicate<EntityType<?>> velocityUpdateSupplier = EntityType::defaultTrackDeltasSupplier;
+        private java.util.function.ToIntFunction<EntityType<?>> trackingRangeSupplier = EntityType::defaultTrackingRangeSupplier;
+        private java.util.function.ToIntFunction<EntityType<?>> updateIntervalSupplier = EntityType::defaultUpdateIntervalSupplier;
+        private boolean onlyOpCanSetNbt = false;
 
         private Builder(EntityType.EntityFactory<T> factory, MobCategory category) {
             this.factory = factory;
@@ -1644,7 +_,7 @@
             this.canSpawnFarFromPlayer = category == MobCategory.CREATURE || category == MobCategory.MISC;
         }
 
-        public static <T extends Entity> EntityType.Builder<T> of(EntityType.EntityFactory<T> factory, MobCategory category) {
+        public static <T extends Entity> EntityType.Builder<T> of(EntityType.EntityFactory factory, MobCategory category) { // CraftBukkit - decompile error
             return new EntityType.Builder<>(factory, category);
         }
 
@@ -1755,6 +_,26 @@
             return this;
         }
 
+        public EntityType.Builder<T> setUpdateInterval(int interval) {
+            this.updateIntervalSupplier = t->interval;
+            return this;
+        }
+
+        public EntityType.Builder<T> setTrackingRange(int range) {
+            this.trackingRangeSupplier = t->range;
+            return this;
+        }
+
+        public EntityType.Builder<T> setShouldReceiveVelocityUpdates(boolean value) {
+            this.velocityUpdateSupplier = t->value;
+            return this;
+        }
+
+        public EntityType.Builder<T> setOnlyOpCanSetNbt(boolean onlyOpCanSetNbt) {
+            this.onlyOpCanSetNbt = onlyOpCanSetNbt;
+            return this;
+        }
+
         public EntityType<T> build(ResourceKey<EntityType<?>> name) {
             if (this.serialize) {
                 Util.fetchChoiceType(References.ENTITY_TREE, name.identifier().toString());
@@ -1775,7 +_,11 @@
                 this.descriptionId.get(name),
                 this.lootTable.get(name),
                 this.requiredFeatures,
-                this.allowedInPeaceful
+                this.allowedInPeaceful,
+                this.velocityUpdateSupplier,
+                this.trackingRangeSupplier,
+                this.updateIntervalSupplier,
+                this.onlyOpCanSetNbt
             );
         }
     }
