--- a/net/minecraft/world/entity/EntitySelector.java
+++ b/net/minecraft/world/entity/EntitySelector.java
@@ -33,7 +_,7 @@
             ? Predicates.alwaysFalse()
             : NO_SPECTATORS.and(
                 input -> {
-                    if (!input.isPushable()) {
+                    if (!input.canCollideWithBukkit(entity) || !entity.canCollideWithBukkit(input)) { // CraftBukkit - collidable API
                         return false;
                     } else if (!entity.level().isClientSide() || input instanceof Player player && player.isLocalPlayer()) {
                         Team theirTeam = input.getTeam();
