--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -2,6 +_,7 @@
 
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.protocol.Packet;
@@ -23,6 +_,8 @@
 import net.minecraft.world.level.entity.EntityTypeTest;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class ExperienceOrb extends Entity {
     private static final int LIFETIME = 6000;
@@ -33,11 +_,64 @@
     private int age;
     private int health = 5;
     public int value;
-    private int count = 1;
+    public int count = 1;
     private Player followingPlayer;
 
+    // Paper start
+    @javax.annotation.Nullable
+    public java.util.UUID sourceEntityId;
+    @javax.annotation.Nullable
+    public java.util.UUID triggerEntityId;
+    public org.bukkit.entity.ExperienceOrb.SpawnReason spawnReason = org.bukkit.entity.ExperienceOrb.SpawnReason.UNKNOWN;
+
+    private void loadPaperNBT(CompoundTag tag) {
+        if (!tag.contains("Paper.ExpData", net.minecraft.nbt.Tag.TAG_COMPOUND)) {
+            return;
+        }
+        CompoundTag comp = tag.getCompound("Paper.ExpData");
+        if (comp.hasUUID("source")) {
+            this.sourceEntityId = comp.getUUID("source");
+        }
+        if (comp.hasUUID("trigger")) {
+            this.triggerEntityId = comp.getUUID("trigger");
+        }
+        if (comp.contains("reason")) {
+            String reason = comp.getString("reason");
+            try {
+                this.spawnReason = org.bukkit.entity.ExperienceOrb.SpawnReason.valueOf(reason);
+            } catch (Exception e) {
+                this.level().getCraftServer().getLogger().warning("Invalid spawnReason set for experience orb: " + e.getMessage() + " - " + reason);
+            }
+        }
+    }
+    private void savePaperNBT(CompoundTag tag) {
+        CompoundTag comp = new CompoundTag();
+        if (this.sourceEntityId != null) {
+            comp.putUUID("source", this.sourceEntityId);
+        }
+        if (this.triggerEntityId != null) {
+            comp.putUUID("trigger", triggerEntityId);
+        }
+        if (this.spawnReason != null && this.spawnReason != org.bukkit.entity.ExperienceOrb.SpawnReason.UNKNOWN) {
+            comp.putString("reason", this.spawnReason.name());
+        }
+        tag.put("Paper.ExpData", comp);
+    }
+
     public ExperienceOrb(Level p_20776_, double p_20777_, double p_20778_, double p_20779_, int p_20780_) {
+        this(p_20776_, p_20777_, p_20778_, p_20779_, p_20780_, null, null);
+    }
+
+    public ExperienceOrb(Level world, double x, double y, double z, int amount, @javax.annotation.Nullable org.bukkit.entity.ExperienceOrb.SpawnReason reason, @javax.annotation.Nullable Entity triggerId) {
+        this(world, x, y, z, amount, reason, triggerId, null);
+    }
+
+    public ExperienceOrb(Level p_20776_, double p_20777_, double p_20778_, double p_20779_, int p_20780_, @javax.annotation.Nullable org.bukkit.entity.ExperienceOrb.SpawnReason reason, @javax.annotation.Nullable Entity triggerId, @javax.annotation.Nullable Entity sourceId) {
         this(EntityType.EXPERIENCE_ORB, p_20776_);
+        this.sourceEntityId = sourceId != null ? sourceId.getUUID() : null;
+        this.triggerEntityId = triggerId != null ? triggerId.getUUID() : null;
+        this.spawnReason = reason != null ? reason : org.bukkit.entity.ExperienceOrb.SpawnReason.UNKNOWN;
+        // Paper end
         this.setPos(p_20777_, p_20778_, p_20779_);
         this.setYRot((float)(this.random.nextDouble() * 360.0));
         this.setDeltaMovement(
@@ -67,6 +_,7 @@
     @Override
     public void tick() {
         super.tick();
+        Player prevTarget = this.followingPlayer;// CraftBukkit - store old target
         this.xo = this.getX();
         this.yo = this.getY();
         this.zo = this.getZ();
@@ -96,7 +_,22 @@
             this.followingPlayer = null;
         }
 
-        if (this.followingPlayer != null) {
+        // CraftBukkit start
+        boolean cancelled = false;
+        if (this.followingPlayer != prevTarget) {
+            org.bukkit.event.entity.EntityTargetLivingEntityEvent event = CraftEventFactory.callEntityTargetLivingEvent(this, followingPlayer, (followingPlayer != null) ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET);
+            LivingEntity target = (event.getTarget() == null) ? null : ((CraftLivingEntity) event.getTarget()).getHandle();
+            cancelled = event.isCancelled();
+
+            if (cancelled) {
+                followingPlayer = prevTarget;
+            } else {
+                followingPlayer = (target instanceof Player) ? (Player) target : null;
+            }
+        }
+
+        if (this.followingPlayer != null && !cancelled) {
+        // CraftBukkit end
             Vec3 vec3 = new Vec3(
                 this.followingPlayer.getX() - this.getX(),
                 this.followingPlayer.getY() + (double)this.followingPlayer.getEyeHeight() / 2.0 - this.getY(),
@@ -112,7 +_,8 @@
         this.move(MoverType.SELF, this.getDeltaMovement());
         float f = 0.98F;
         if (this.onGround()) {
-            f = this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.98F;
+            BlockPos pos = getBlockPosBelowThatAffectsMyMovement();
+            f = this.level().getBlockState(pos).getFriction(this.level(), pos, this) * 0.98F;
         }
 
         this.setDeltaMovement(this.getDeltaMovement().multiply((double)f, 0.98, (double)f));
@@ -122,7 +_,7 @@
 
         this.age++;
         if (this.age >= 6000) {
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         }
     }
 
@@ -133,7 +_,7 @@
 
     private void scanForEntities() {
         if (this.followingPlayer == null || this.followingPlayer.distanceToSqr(this) > 64.0) {
-            this.followingPlayer = this.level().getNearestPlayer(this, 8.0);
+            this.followingPlayer = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.entity.XpOrbTargetingEvent(this, 8.0)).getFollowingPlayer();
         }
 
         if (this.level() instanceof ServerLevel) {
@@ -144,12 +_,36 @@
         }
     }
 
+    private static AtomicReference<org.bukkit.entity.ExperienceOrb.SpawnReason> award$reason = new AtomicReference<>(null);
+    private static AtomicReference<Entity> award$triggerId = new AtomicReference<>(null);
+    private static AtomicReference<Entity> award$sourceId = new AtomicReference<>(null);
+
+    public static void awardPaper(org.bukkit.entity.ExperienceOrb.SpawnReason reason, Entity triggerId, Entity sourceId) {
+        award$reason.set(reason);
+        award$triggerId.set(triggerId);
+        award$sourceId.set(sourceId);
+    }
+
     public static void award(ServerLevel p_147083_, Vec3 p_147084_, int p_147085_) {
-        while (p_147085_ > 0) {
-            int i = getExperienceValue(p_147085_);
-            p_147085_ -= i;
-            if (!tryMergeToExisting(p_147083_, p_147084_, i)) {
-                p_147083_.addFreshEntity(new ExperienceOrb(p_147083_, p_147084_.x(), p_147084_.y(), p_147084_.z(), i));
+        // Paper end - add reasons for orbs
+        while (p_147085_ > 0) {
+            int i = getExperienceValue(p_147085_);
+            p_147085_ -= i;
+            if (!tryMergeToExisting(p_147083_, p_147084_, i)) {
+                p_147083_.addFreshEntity(new ExperienceOrb(p_147083_, p_147084_.x(), p_147084_.y(), p_147084_.z(), i, award$reason.getAndSet(null), award$triggerId.getAndSet(null), award$sourceId.getAndSet(null))); // Paper - add reason
+            }
+        }
+    }
+    public static void award(ServerLevel world, Vec3 pos, int amount, org.bukkit.entity.ExperienceOrb.SpawnReason reason, Entity triggerId) {
+        award(world, pos, amount, reason, triggerId, null);
+    }
+    public static void award(ServerLevel p_147083_, Vec3 p_147084_, int p_147085_, org.bukkit.entity.ExperienceOrb.SpawnReason reason, Entity triggerId, Entity sourceId) {
+        // Paper end - add reasons for orbs
+        while (p_147085_ > 0) {
+            int i = getExperienceValue(p_147085_);
+            p_147085_ -= i;
+            if (!tryMergeToExisting(p_147083_, p_147084_, i)) {
+                p_147083_.addFreshEntity(new ExperienceOrb(p_147083_, p_147084_.x(), p_147084_.y(), p_147084_.z(), i, reason, triggerId, sourceId)); // Paper - add reason
             }
         }
     }
@@ -177,8 +_,14 @@
     }
 
     private void merge(ExperienceOrb p_147101_) {
+        // Paper start - call orb merge event
+        if (!new com.destroystokyo.paper.event.entity.ExperienceOrbMergeEvent((org.bukkit.entity.ExperienceOrb) this.getBukkitEntity(), (org.bukkit.entity.ExperienceOrb) p_147101_.getBukkitEntity()).callEvent()) {
+            return;
+        }
+        // Paper end - call orb merge event
         this.count = this.count + p_147101_.count;
         this.age = Math.min(this.age, p_147101_.age);
+        p_147101_.setRemoved$cause(org.bukkit.event.entity.EntityRemoveEvent.Cause.MERGE);  // CraftBukkit - add Bukkit remove cause
         p_147101_.discard();
     }
 
@@ -201,7 +_,7 @@
             this.markHurt();
             this.health = (int)((float)this.health - p_20786_);
             if (this.health <= 0) {
-                this.discard();
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause
             }
 
             return true;
@@ -212,32 +_,36 @@
     public void addAdditionalSaveData(CompoundTag p_20796_) {
         p_20796_.putShort("Health", (short)this.health);
         p_20796_.putShort("Age", (short)this.age);
-        p_20796_.putShort("Value", (short)this.value);
+        p_20796_.putInt("Value", this.value);  // Paper - save as Integer
         p_20796_.putInt("Count", this.count);
+        this.savePaperNBT(p_20796_); // Paper
     }
 
     @Override
     public void readAdditionalSaveData(CompoundTag p_20788_) {
         this.health = p_20788_.getShort("Health");
         this.age = p_20788_.getShort("Age");
-        this.value = p_20788_.getShort("Value");
+        this.value = p_20788_.getInt("Value"); // Paper - load as Integer
         this.count = Math.max(p_20788_.getInt("Count"), 1);
+        this.loadPaperNBT(p_20788_); // Paper
     }
 
     @Override
     public void playerTouch(Player p_20792_) {
         if (p_20792_ instanceof ServerPlayer serverplayer) {
-            if (p_20792_.takeXpDelay == 0) {
-                p_20792_.takeXpDelay = 2;
+            if (p_20792_.takeXpDelay == 0 && new com.destroystokyo.paper.event.player.PlayerPickupExperienceEvent(serverplayer.getBukkitEntity(), (org.bukkit.entity.ExperienceOrb) this.getBukkitEntity()).callEvent()) { // Paper - PlayerPickupExperienceEvent
+                if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.PlayerXpEvent.PickupXp(p_20792_, this)).isCanceled()) return;
+                p_20792_.takeXpDelay = CraftEventFactory.callPlayerXpCooldownEvent(p_20792_, 2, org.bukkit.event.player.PlayerExpCooldownChangeEvent.ChangeReason.PICKUP_ORB).getNewCooldown(); // CraftBukkit - p_20792_.takeXpDelay = 2;
                 p_20792_.take(this, 1);
                 int i = this.repairPlayerItems(serverplayer, this.value);
+
                 if (i > 0) {
-                    p_20792_.giveExperiencePoints(i);
+                    p_20792_.giveExperiencePoints(CraftEventFactory.callPlayerExpChangeEvent(p_20792_, this).getAmount()); // CraftBukkit - this.value -> event.getAmount()
                 }
 
                 this.count--;
                 if (this.count == 0) {
-                    this.discard();
+                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
                 }
             }
         }
@@ -247,8 +_,19 @@
         Optional<EnchantedItemInUse> optional = EnchantmentHelper.getRandomItemWith(EnchantmentEffectComponents.REPAIR_WITH_XP, p_344821_, ItemStack::isDamaged);
         if (optional.isPresent()) {
             ItemStack itemstack = optional.get().itemStack();
-            int i = EnchantmentHelper.modifyDurabilityToRepairFromXp(p_344821_.serverLevel(), itemstack, p_147094_);
+            int i = EnchantmentHelper.modifyDurabilityToRepairFromXp(p_344821_.serverLevel(), itemstack, (int) (p_147094_ * itemstack.getXpRepairRatio()));
             int j = Math.min(i, itemstack.getDamageValue());
+            // CraftBukkit start
+            // Paper start - mending event
+            final int consumedExperience = j > 0 ? j * p_147094_ / j : 0;
+            org.bukkit.event.player.PlayerItemMendEvent event = CraftEventFactory.callPlayerItemMendEvent(p_344821_, this, itemstack, optional.get().inSlot(), j, consumedExperience);
+            // Paper end - mending event
+            j = event.getRepairAmount();
+            if (event.isCancelled()) {
+                return p_147094_;
+            }
+            // CraftBukkit end
+
             itemstack.setDamageValue(itemstack.getDamageValue() - j);
             if (j > 0) {
                 int k = p_147094_ - j * p_147094_ / i;
@@ -292,6 +_,25 @@
     }
 
     public static int getExperienceValue(int p_20783_) {
+        // CraftBukkit start
+        if (p_20783_ > 162670129) return p_20783_ - 100000;
+        if (p_20783_ > 81335063) return 81335063;
+        if (p_20783_ > 40667527) return 40667527;
+        if (p_20783_ > 20333759) return 20333759;
+        if (p_20783_ > 10166857) return 10166857;
+        if (p_20783_ > 5083423) return 5083423;
+        if (p_20783_ > 2541701) return 2541701;
+        if (p_20783_ > 1270849) return 1270849;
+        if (p_20783_ > 635413) return 635413;
+        if (p_20783_ > 317701) return 317701;
+        if (p_20783_ > 158849) return 158849;
+        if (p_20783_ > 79423) return 79423;
+        if (p_20783_ > 39709) return 39709;
+        if (p_20783_ > 19853) return 19853;
+        if (p_20783_ > 9923) return 9923;
+        if (p_20783_ > 4957) return 4957;
+        // CraftBukkit end
+
         if (p_20783_ >= 2477) {
             return 2477;
         } else if (p_20783_ >= 1237) {
