--- a/net/minecraft/world/entity/Interaction.java
+++ b/net/minecraft/world/entity/Interaction.java
@@ -205,7 +_,7 @@
         return this.getDimensions().makeBoundingBox(this.position());
     }
 
-    static record PlayerAction(UUID player, long timestamp) {
+    public static record PlayerAction(UUID player, long timestamp) {
         public static final Codec<Interaction.PlayerAction> CODEC = RecordCodecBuilder.create(
             p_273237_ -> p_273237_.group(
                         UUIDUtil.CODEC.fieldOf("player").forGetter(Interaction.PlayerAction::player),
