--- a/net/minecraft/world/entity/InterpolationTracker.java
+++ b/net/minecraft/world/entity/InterpolationTracker.java
@@ -3,26 +_,9 @@
 import net.minecraft.world.phys.Vec3;
 
 public interface InterpolationTracker {
-    InterpolationTracker NO_OP = new InterpolationTracker.NoOpInterpolationTracker();
-
     void updateTracking(Vec3 trackingPos);
 
     PositionPath getPositionPath(Vec3 trackingPos);
 
     void clear();
-
-    record NoOpInterpolationTracker() implements InterpolationTracker {
-        @Override
-        public void updateTracking(Vec3 trackingPos) {
-        }
-
-        @Override
-        public PositionPath getPositionPath(Vec3 trackingPos) {
-            return PositionPath.of(trackingPos);
-        }
-
-        @Override
-        public void clear() {
-        }
-    }
 }
