--- a/net/minecraft/world/entity/ItemBasedSteering.java
+++ b/net/minecraft/world/entity/ItemBasedSteering.java
@@ -12,8 +_,8 @@
     private final SynchedEntityData entityData;
     private final EntityDataAccessor<Integer> boostTimeAccessor;
     private final EntityDataAccessor<Boolean> hasSaddleAccessor;
-    public boolean boosting;
-    public int boostTime;
+    private boolean boosting;
+    private int boostTime;
 
     public ItemBasedSteering(SynchedEntityData p_20841_, EntityDataAccessor<Integer> p_20842_, EntityDataAccessor<Boolean> p_20843_) {
         this.entityData = p_20841_;
@@ -47,7 +_,7 @@
         return this.boosting ? 1.0F + 1.15F * Mth.sin((float)this.boostTime / (float)this.boostTimeTotal() * (float) Math.PI) : 1.0F;
     }
 
-    public int boostTimeTotal() {
+    private int boostTimeTotal() {
         return this.entityData.get(this.boostTimeAccessor);
     }
 
