--- a/net/minecraft/world/entity/ItemOwner.java
+++ b/net/minecraft/world/entity/ItemOwner.java
@@ -19,7 +_,7 @@
         return new ItemOwner.OffsetFromOwner(owner, offset);
     }
 
-    record OffsetFromOwner(ItemOwner owner, Vec3 offset) implements ItemOwner {
+    public record OffsetFromOwner(ItemOwner owner, Vec3 offset) implements ItemOwner {
         @Override
         public Level level() {
             return this.owner.level();
