--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -23,6 +_,7 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityUnleashEvent;
 import org.jspecify.annotations.Nullable;
 
 public interface Leashable {
@@ -83,6 +_,14 @@
     }
 
     default void writeLeashData(ValueOutput output, Leashable.@Nullable LeashData leashData) {
+        // CraftBukkit start - SPIGOT-7487: Don't save (and possible drop) leash, when the holder was removed by a plugin
+        if (leashData != null) {
+            Entity entity = leashData.leashHolder;
+            if (entity != null && entity.pluginRemoved) {
+                return;
+            }
+        }
+        // CraftBukkit end
         output.storeNullable("leash", Leashable.LeashData.CODEC, leashData);
     }
 
@@ -146,7 +_,15 @@
 
         if (leashData != null && leashData.leashHolder != null) {
             if (!entity.canInteractWithLevel() || !leashData.leashHolder.canInteractWithLevel()) {
-                if (level.getGameRules().get(GameRules.ENTITY_DROPS)) {
+                // Paper start - Expand EntityUnleashEvent
+                final org.bukkit.event.entity.EntityUnleashEvent event = new org.bukkit.event.entity.EntityUnleashEvent(
+                    entity.getBukkitEntity(),
+                    !entity.isAlive() ? EntityUnleashEvent.UnleashReason.LEASHED_GONE : EntityUnleashEvent.UnleashReason.HOLDER_GONE,
+                    level.getGameRules().get(GameRules.ENTITY_DROPS) && !entity.pluginRemoved
+                );
+                event.callEvent();
+                if (event.isDropLeash()) { // CraftBukkit - SPIGOT-7487: Don't drop leash, when the holder was removed by a plugin
+                    // Paper end - Expand EntityUnleashEvent
                     entity.dropLeash();
                 } else {
                     entity.removeLeash();
@@ -157,8 +_,7 @@
             if (leashHolder != null && leashHolder.level() == entity.level()) {
                 double distanceTo = entity.leashDistanceTo(leashHolder);
                 entity.whenLeashedTo(leashHolder);
-                if (distanceTo > entity.leashSnapDistance()) {
-                    level.playSound(null, leashHolder.getX(), leashHolder.getY(), leashHolder.getZ(), SoundEvents.LEAD_BREAK, SoundSource.NEUTRAL, 1.0F, 1.0F);
+                if (distanceTo > entity.leashSnapDistanceOrConfig()) { // Paper - Configurable max leash distance
                     entity.leashTooFarBehaviour();
                 } else if (distanceTo > entity.leashElasticDistance() - leashHolder.getBbWidth() - entity.getBbWidth()
                     && entity.checkElasticInteractions(leashHolder, leashData)) {
@@ -178,6 +_,12 @@
         entity.checkFallDistanceAccumulation();
     }
 
+    // Paper start - Configurable max leash distance
+    default double leashSnapDistanceOrConfig() {
+        if (!(this instanceof final Entity entity)) return leashSnapDistance();
+        return entity.level().paperConfig().misc.maxLeashDistance.or(leashSnapDistance());
+    }
+    // Paper end - Configurable max leash distance
     default double leashSnapDistance() {
         return 12.0;
     }
@@ -188,7 +_,8 @@
 
     static <E extends Entity & Leashable> float angularFriction(E entity) {
         if (entity.onGround()) {
-            return entity.level().getBlockState(entity.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction() * 0.91F;
+            var groundPos = entity.getBlockPosBelowThatAffectsMyMovement();
+            return entity.level().getBlockState(groundPos).getFriction(entity.level(), groundPos, entity) * 0.91F;
         } else {
             return entity.isInLiquid() ? 0.8F : 0.91F;
         }
@@ -199,7 +_,25 @@
     }
 
     default void leashTooFarBehaviour() {
-        this.dropLeash();
+        // CraftBukkit start
+        boolean dropLeash = true; // Paper
+        if (this instanceof Entity entity) {
+            // Paper start - Expand EntityUnleashEvent
+            final org.bukkit.event.entity.EntityUnleashEvent event = new org.bukkit.event.entity.EntityUnleashEvent(entity.getBukkitEntity(), EntityUnleashEvent.UnleashReason.DISTANCE, true);
+            if (!event.callEvent()) return;
+
+            Entity leashHolder = this.getLeashHolder();
+            Level level = leashHolder.level();
+            dropLeash = event.isDropLeash();
+            level.playSound(null, leashHolder.getX(), leashHolder.getY(), leashHolder.getZ(), SoundEvents.LEAD_BREAK, SoundSource.NEUTRAL, 1.0F, 1.0F); // Moved from Leashable#tickLeash
+        }
+        // CraftBukkit end
+        if (dropLeash) {
+            this.dropLeash();
+        } else {
+            this.removeLeash();
+        }
+        // Paper end - Expand EntityUnleashEvent
     }
 
     default void closeRangeLeashBehaviour(Entity leashHolder) {
@@ -362,7 +_,7 @@
             .toList();
     }
 
-    final class LeashData {
+    public static final class LeashData {
         public static final Codec<Leashable.LeashData> CODEC = Codec.xor(UUIDUtil.CODEC.fieldOf("UUID").codec(), BlockPos.CODEC)
             .xmap(
                 Leashable.LeashData::new,
