--- a/net/minecraft/world/entity/LightningBolt.java
+++ b/net/minecraft/world/entity/LightningBolt.java
@@ -25,6 +_,7 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.v1_21_R1.event.CraftEventFactory;
 
 public class LightningBolt extends Entity {
     private static final int START_LIFE = 2;
@@ -38,6 +_,7 @@
     private ServerPlayer cause;
     private final Set<Entity> hitEntities = Sets.newHashSet();
     private int blocksSetOnFire;
+    private float damage = 5.0F;
 
     public LightningBolt(EntityType<? extends LightningBolt> p_20865_, Level p_20866_) {
         super(p_20865_, p_20866_);
@@ -73,6 +_,14 @@
         }
     }
 
+    public void setDamage(float damage) {
+        this.damage = damage;
+    }
+
+    public float getDamage() {
+        return this.damage;
+    }
+
     @Override
     public void tick() {
         super.tick();
@@ -130,7 +_,7 @@
                     }
                 }
 
-                this.discard();
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             } else if (this.life < -this.random.nextInt(10)) {
                 this.flashes--;
                 this.life = 1;
@@ -139,7 +_,7 @@
             }
         }
 
-        if (this.life >= 0) {
+        if (this.life >= 0 && !this.visualOnly) { // CraftBukkit - add !this.visualOnly
             if (!(this.level() instanceof ServerLevel)) {
                 this.level().setSkyFlashTime(2);
             } else if (!this.visualOnly) {
@@ -151,6 +_,7 @@
                     );
 
                 for (Entity entity : list1) {
+                    if (!net.neoforged.neoforge.event.EventHooks.onEntityStruckByLightning(entity, this))
                     entity.thunderHit((ServerLevel)this.level(), this);
                 }
 
@@ -172,16 +_,24 @@
             BlockPos blockpos = this.blockPosition();
             BlockState blockstate = BaseFireBlock.getState(this.level(), blockpos);
             if (this.level().getBlockState(blockpos).isAir() && blockstate.canSurvive(this.level(), blockpos)) {
-                this.level().setBlockAndUpdate(blockpos, blockstate);
-                this.blocksSetOnFire++;
+                // CraftBukkit start - add "!visualOnly"
+                if (!visualOnly && !CraftEventFactory.callBlockIgniteEvent(this.level(), blockpos, this).isCancelled()) {
+                    this.level().setBlockAndUpdate(blockpos, blockstate);
+                    ++this.blocksSetOnFire;
+                }
+                // CraftBukkit end
             }
 
             for (int i = 0; i < p_20871_; i++) {
                 BlockPos blockpos1 = blockpos.offset(this.random.nextInt(3) - 1, this.random.nextInt(3) - 1, this.random.nextInt(3) - 1);
                 blockstate = BaseFireBlock.getState(this.level(), blockpos1);
                 if (this.level().getBlockState(blockpos1).isAir() && blockstate.canSurvive(this.level(), blockpos1)) {
-                    this.level().setBlockAndUpdate(blockpos1, blockstate);
-                    this.blocksSetOnFire++;
+                    // CraftBukkit start - add "!visualOnly"
+                    if (!visualOnly && !CraftEventFactory.callBlockIgniteEvent(this.level(), blockpos1, this).isCancelled()) {
+                        this.level().setBlockAndUpdate(blockpos1, blockstate);
+                        ++this.blocksSetOnFire;
+                    }
+                    // CraftBukkit end
                 }
             }
         }
