--- a/net/minecraft/world/entity/LinearInterpolationHandler.java
+++ b/net/minecraft/world/entity/LinearInterpolationHandler.java
@@ -4,20 +_,15 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.phys.Vec3;
 
-public class LinearInterpolationHandler extends AbstractInterpolationHandler {
-    public static final int DEFAULT_INTERPOLATION_STEPS = 3;
+public class LinearInterpolationHandler extends InterpolationHandler {
     private final LinearInterpolationHandler.InterpolationData interpolationData = new LinearInterpolationHandler.InterpolationData();
 
-    private LinearInterpolationHandler(Entity entity, int interpolationSteps) {
+    public LinearInterpolationHandler(Entity entity) {
+        super(entity);
+    }
+
+    public LinearInterpolationHandler(Entity entity, int interpolationSteps) {
         super(entity, interpolationSteps);
-    }
-
-    public static InterpolationHandler create(Entity entity, int interpolationSteps) {
-        return entity.level().isClientSide() ? new LinearInterpolationHandler(entity, interpolationSteps) : InterpolationHandler.NO_OP;
-    }
-
-    public static InterpolationHandler create(Entity entity) {
-        return create(entity, 3);
     }
 
     public void setInterpolationLength(int steps) {
