--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -5,6 +_,7 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
+import com.mohistmc.youer.neoforge.BukkitDamageHooks;
 import com.mojang.datafixers.util.Pair;
 import com.mojang.logging.LogUtils;
 import it.unimi.dsi.fastutil.doubles.DoubleDoubleImmutablePair;
@@ -12,9 +_,11 @@
 import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap;
 import it.unimi.dsi.fastutil.objects.Reference2ObjectArrayMap;
 import it.unimi.dsi.fastutil.objects.Reference2ObjectMap;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.ConcurrentModificationException;
 import java.util.EnumMap;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
@@ -22,6 +_,8 @@
 import java.util.Set;
 import java.util.UUID;
 import java.util.Map.Entry;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BiConsumer;
 import java.util.function.Consumer;
 import java.util.function.Function;
@@ -98,6 +_,7 @@
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.component.AttackRange;
 import net.minecraft.world.item.component.BlocksAttacks;
+import net.minecraft.world.item.component.Consumable;
 import net.minecraft.world.item.component.DeathProtection;
 import net.minecraft.world.item.component.KineticWeapon;
 import net.minecraft.world.item.component.Weapon;
@@ -134,6 +_,15 @@
 import net.minecraft.world.scores.Scoreboard;
 import net.minecraft.world.waypoints.Waypoint;
 import net.minecraft.world.waypoints.WaypointTransmitter;
+import org.bukkit.craftbukkit.attribute.CraftAttributeMap;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.ArrowBodyCountChangeEvent;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+import org.bukkit.event.entity.EntityRegainHealthEvent;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.player.PlayerDropItemEvent;
 import org.jetbrains.annotations.Contract;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
@@ -211,7 +_,8 @@
     public static final float ELYTRA_VERTICAL_AIR_DRAG = 0.98F;
     public static final float BASE_SWIM_SPEED = 0.02F;
     private int currentImpulseContextResetGraceTime = 0;
-    public static final Predicate<LivingEntity> PLAYER_NOT_WEARING_DISGUISE_ITEM = livingEntity -> {
+    // Neo: Support IItemExtension#isGazeDisguise
+    public static final java.util.function.BiPredicate<LivingEntity, @org.jspecify.annotations.Nullable LivingEntity> PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET = (livingEntity, target) -> {
         if (livingEntity instanceof Player player) {
             ItemStack helmet = player.getItemBySlot(EquipmentSlot.HEAD);
             return !helmet.is(ItemTags.GAZE_DISGUISE_EQUIPMENT);
@@ -219,6 +_,9 @@
             return true;
         }
     };
+    /** @deprecated Neo: use {@link #PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET} with target info instead */
+    @Deprecated
+    public static final Predicate<LivingEntity> PLAYER_NOT_WEARING_DISGUISE_ITEM = livingEntity -> PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET.test(livingEntity, null);
     private final AttributeMap attributes;
     public CombatTracker combatTracker = new CombatTracker(this);
     public final Map<Holder<MobEffect>, MobEffectInstance> activeEffects = Maps.newHashMap();
@@ -281,14 +_,40 @@
     private final EnumMap<EquipmentSlot, Reference2ObjectMap<Enchantment, Set<EnchantmentLocationBasedEffect>>> activeLocationDependentEnchantments = new EnumMap<>(
         EquipmentSlot.class
     );
-    protected final EntityEquipment equipment;
+    public final EntityEquipment equipment;
     private Waypoint.Icon locatorBarIcon = new Waypoint.Icon();
     public @Nullable Vec3 currentImpulseImpactPos;
+    /**
+     * This field stores information about damage dealt to this entity.
+     * a new {@link net.neoforged.neoforge.common.damagesource.DamageContainer} is instantiated
+     * via {@link #hurt(DamageSource, float)} after invulnerability checks, and is removed from
+     * the stack before the method's return.
+     **/
+    protected java.util.@Nullable Stack<net.neoforged.neoforge.common.damagesource.DamageContainer> damageContainers = new java.util.Stack<>();
+
+    // CraftBukkit start
+    public int expToDrop;
+    public @Nullable List<DefaultDrop> deathDropItems = null;
+    public final org.bukkit.craftbukkit.attribute.CraftAttributeMap craftAttributes;
+    public boolean collides = true;
+    public Set<UUID> collidableExemptions = new java.util.HashSet<>();
+    public boolean bukkitPickUpLoot;
+    public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
+    public boolean silentDeath = true; // Paper - mark entity as dying silently for cancellable death event
+    public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper - Friction API
+    public int invulnerableDuration = LivingEntity.INVULNERABLE_DURATION; // Paper - configurable invulnerable duration
+    // CraftBukkit end
+
+    // CraftBukkit end
+    public EntityDamageEvent bukkitEntityDamageEvent;
 
     protected LivingEntity(EntityType<? extends LivingEntity> type, Level level) {
         super(type, level);
         this.attributes = new AttributeMap(DefaultAttributes.getSupplier(type));
+        this.craftAttributes = new CraftAttributeMap(attributes); // CraftBukkit
+        this.fake$setHealth(true);
         this.setHealth(this.getMaxHealth());
+        this.entityData.set(DATA_HEALTH_ID, (float) this.getAttribute(Attributes.MAX_HEALTH).getValue());
         this.equipment = this.createEquipment();
         this.blocksBuilding = true;
         this.reapplyPosition();
@@ -358,12 +_,15 @@
             .add(Attributes.AIR_DRAG_MODIFIER)
             .add(Attributes.FRICTION_MODIFIER)
             .add(Attributes.NAME_TAG_DISTANCE)
+            .add(net.neoforged.neoforge.common.NeoForgeMod.SWIM_SPEED)
+            .add(net.neoforged.neoforge.common.NeoForgeMod.NAMETAG_DISTANCE)
             .add(Attributes.BELOW_NAME_DISTANCE);
     }
 
     @Override
     protected void checkFallDamage(double ya, boolean onGround, BlockState onState, BlockPos pos) {
         if (!this.isInWater()) {
+            // TODO: prior to 26.1-snapshot-8 we called our patched in method 'this.updateInWaterStateAndDoWaterCurrentPushing(false)'
             this.updateFluidInteraction();
         }
 
@@ -385,7 +_,15 @@
 
                 double scale = Math.min(0.2F + power / 15.0, 2.5);
                 int particles = (int)(150.0 * scale);
-                level.sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, onState), x, y, z, particles, 0.0, 0.0, 0.0, 0.15F);
+                if (!onState.addLandingEffects(level, pos, onState, this, particles)) {
+                    // CraftBukkit start - visiblity api
+                    if (this instanceof ServerPlayer) {
+                        level.sendParticlesSource((ServerPlayer) this, new BlockParticleOption(ParticleTypes.BLOCK, onState, pos), false, false, x, y, z, particles, 0.0, 0.0, 0.0, 0.15F);
+                    } else {
+                        level.sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, onState, pos), x, y, z, particles, 0.0, 0.0, 0.0, 0.15F);
+                    }
+                    // CraftBukkit end
+                }
             }
         }
 
@@ -395,6 +_,7 @@
         }
     }
 
+    @Deprecated //FORGE: Use canDrownInFluidType instead
     public boolean canBreatheUnderwater() {
         return this.is(EntityTypeTags.CAN_BREATHE_UNDER_WATER);
     }
@@ -574,7 +_,7 @@
         this.deathTime++;
         if (this.deathTime >= 20 && !this.level().isClientSide() && !this.isRemoved()) {
             this.level().broadcastEntityEvent(this, (byte)60);
-            this.remove(Entity.RemovalReason.KILLED);
+            this.remove(Entity.RemovalReason.KILLED, EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause
         }
     }
 
@@ -688,10 +_,16 @@
     }
 
     public void onEquipItem(EquipmentSlot slot, ItemStack oldStack, ItemStack stack) {
+        // CraftBukkit start
+        onEquipItem(slot, oldStack, stack, false);
+    }
+
+    public void onEquipItem(EquipmentSlot slot, ItemStack oldStack, ItemStack stack, boolean silent) {
+        // CraftBukkit end
         if (!this.level().isClientSide() && !this.isSpectator()) {
             if (!ItemStack.isSameItemSameComponents(oldStack, stack) && !this.firstTick) {
                 Equippable equippable = stack.get(DataComponents.EQUIPPABLE);
-                if (!this.isSilent() && equippable != null && slot == equippable.slot()) {
+                if (!this.isSilent() && equippable != null && slot == equippable.slot() && !silent) { // CraftBukkit
                     this.level()
                         .playSeededSound(
                             null,
@@ -719,11 +_,18 @@
 
     @Override
     public void remove(Entity.RemovalReason reason) {
+        // CraftBukkit start - add Bukkit remove cause
+        this.remove(reason, null);
+    }
+
+    @Override
+    public void remove(Entity.RemovalReason reason, EntityRemoveEvent.Cause cause) {
+        // CraftBukkit end
         if ((reason == Entity.RemovalReason.KILLED || reason == Entity.RemovalReason.DISCARDED) && this.level() instanceof ServerLevel level) {
             this.triggerOnDeathMobEffects(level, reason);
         }
 
-        super.remove(reason);
+        super.remove(reason, cause); // CraftBukkit
         this.brain.clearMemories();
     }
 
@@ -740,6 +_,7 @@
             effect.onMobRemoved(level, this, reason);
         }
 
+        this.removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.DEATH); // CraftBukkit
         this.activeEffects.clear();
     }
 
@@ -778,22 +_,54 @@
         }
     }
 
-    public @Nullable ItemEntity drop(ItemStack itemStack, boolean randomly, boolean thrownFromHand) {
+    // Paper start - Extend dropItem API
+    public final @Nullable ItemEntity drop(ItemStack itemStack, boolean randomly, boolean thrownFromHand) {
+        return this.drop(itemStack, randomly, thrownFromHand, true, null);
+    }
+    public @Nullable ItemEntity drop(ItemStack itemStack, boolean randomly, boolean thrownFromHand, final boolean callEvent, final java.util.function.@Nullable Consumer<org.bukkit.entity.Item> entityOperation) {
+        // Paper end - Extend dropItem API
         if (itemStack.isEmpty()) {
             return null;
-        }
-
-        if (this.level().isClientSide()) {
+        } else if (this.level().isClientSide()) {
             this.swing(InteractionHand.MAIN_HAND);
             return null;
-        }
-
-        ItemEntity entity = this.createItemStackToDrop(itemStack, randomly, thrownFromHand);
-        if (entity != null) {
-            this.level().addFreshEntity(entity);
-        }
-
-        return entity;
+        } else {
+            ItemEntity entity = this.createItemStackToDrop(itemStack, randomly, thrownFromHand);
+            if (entity != null) {
+                if (captureDrops() != null) {
+                    captureDrops().add(entity);
+                } else {
+                    // CraftBukkit start - fire PlayerDropItemEvent
+                    if (callEvent && this instanceof ServerPlayer) {
+                        org.bukkit.entity.Player player = (org.bukkit.entity.Player) this.getBukkitEntity();
+                        org.bukkit.entity.Item drop = (org.bukkit.entity.Item) entity.getBukkitEntity();
+
+                        PlayerDropItemEvent event = new PlayerDropItemEvent(player, drop);
+                        this.level().getCraftServer().getPluginManager().callEvent(event);
+
+                        if (event.isCancelled()) {
+                            org.bukkit.inventory.ItemStack cur = player.getInventory().getItemInHand();
+                            if (thrownFromHand && (cur == null || cur.getAmount() == 0)) {
+                                // The complete stack was dropped
+                                player.getInventory().setItemInHand(drop.getItemStack());
+                            } else if (thrownFromHand && cur.isSimilar(drop.getItemStack()) && cur.getAmount() < cur.getMaxStackSize() && drop.getItemStack().getAmount() == 1) {
+                                // Only one item is dropped
+                                cur.setAmount(cur.getAmount() + 1);
+                                player.getInventory().setItemInHand(cur);
+                            } else {
+                                // Fallback
+                                player.getInventory().addItem(drop.getItemStack());
+                            }
+                            return null;
+                        }
+                    }
+                    // CraftBukkit end
+                    this.level().addFreshEntity(entity);
+                }
+            }
+
+            return entity;
+        }
     }
 
     @Override
@@ -811,9 +_,17 @@
             this.effectsDirty = true;
         }
 
+        // CraftBukkit start
+        float maxHealth = input.getFloatOr("Bukkit.MaxHealth", -1);
+        if (maxHealth > 0) {
+            this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(maxHealth);
+        }
+        // CraftBukkit end
+
         this.setHealth(input.getFloatOr("Health", this.getMaxHealth()));
         this.hurtTime = input.getShortOr("HurtTime", (short)0);
         this.deathTime = input.getShortOr("DeathTime", (short)0);
+        this.lastHurtByMobTimestamp = input.getIntOr("HurtByTimestamp", 0);
         input.getString("Team").ifPresent(teamName -> {
             Scoreboard scoreboard = this.level().getScoreboard();
             PlayerTeam team = scoreboard.getPlayerTeam(teamName);
@@ -847,23 +_,42 @@
         this.updateDirtyEffects();
     }
 
+    private boolean tickingEffects = false; // CraftBukkit
+    private final java.util.Queue<Runnable> deferredEffectUpdates = new java.util.concurrent.ConcurrentLinkedQueue<>(); // CraftBukkit
+
     protected void tickEffects() {
         if (this.level() instanceof ServerLevel serverLevel) {
             Iterator<Holder<MobEffect>> iterator = this.activeEffects.keySet().iterator();
 
+            this.tickingEffects = true; // CraftBukkit
             try {
                 while (iterator.hasNext()) {
                     Holder<MobEffect> mobEffect = iterator.next();
                     MobEffectInstance effect = this.activeEffects.get(mobEffect);
                     if (!effect.tickServer(serverLevel, this, () -> this.onEffectUpdated(effect, true, null))) {
+                        // CraftBukkit start
+                        EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, effect, null, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.EXPIRATION);
+                        if (event.isCancelled()) {
+                            continue;
+                        }
+                        // CraftBukkit end
+                        if (!net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.living.MobEffectEvent.Expired(this, effect)).isCanceled()) {
                         iterator.remove();
                         this.onEffectsRemoved(List.of(effect));
+                        }
                     } else if (effect.getDuration() % 600 == 0) {
                         this.onEffectUpdated(effect, false, null);
                     }
                 }
             } catch (ConcurrentModificationException var6) {
             }
+            // CraftBukkit start
+            this.tickingEffects = false;
+            Runnable task;
+            while ((task = this.deferredEffectUpdates.poll()) != null) {
+                task.run();
+            }
+            // CraftBukkit end
         } else {
             for (MobEffectInstance effect : this.activeEffects.values()) {
                 effect.tickClient();
@@ -904,8 +_,9 @@
         List<ParticleOptions> visibleEffectParticles = this.activeEffects
             .values()
             .stream()
-            .filter(MobEffectInstance::isVisible)
-            .map(MobEffectInstance::getParticleOptions)
+            .map(effect -> net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.living.EffectParticleModificationEvent(this, effect)))
+            .filter(net.neoforged.neoforge.event.entity.living.EffectParticleModificationEvent::isVisible)
+            .map(net.neoforged.neoforge.event.entity.living.EffectParticleModificationEvent::getParticleOptions)
             .toList();
         this.entityData.set(DATA_EFFECT_PARTICLES, visibleEffectParticles);
         this.entityData.set(DATA_EFFECT_AMBIENCE_ID, areAllEffectsAmbient(this.activeEffects.values()));
@@ -944,6 +_,7 @@
             }
         }
 
+        visibilityPercent = net.neoforged.neoforge.common.CommonHooks.getEntityVisibilityMultiplier(this, targetingEntity, visibilityPercent);
         return visibilityPercent;
     }
 
@@ -973,19 +_,32 @@
         this.entityData.set(DATA_EFFECT_PARTICLES, List.of());
     }
 
+    private final AtomicReference<EntityPotionEffectEvent.Cause> removeAllEffects$cause = new AtomicReference<>(EntityPotionEffectEvent.Cause.UNKNOWN);
+    public void removeAllEffects$cause(EntityPotionEffectEvent.Cause cause) {
+        removeAllEffects$cause.set(cause);
+    }
     public boolean removeAllEffects() {
         if (this.level().isClientSide()) {
             return false;
-        }
-
-        if (this.activeEffects.isEmpty()) {
+        } else if (this.activeEffects.isEmpty()) {
             return false;
+        } else {
+            Map<Holder<MobEffect>, MobEffectInstance> map = new java.util.HashMap<>(this.activeEffects.size());
+            for (Map.Entry<Holder<MobEffect>, MobEffectInstance> entry : this.activeEffects.entrySet()) {
+                EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, entry.getValue(), null, removeAllEffects$cause.getAndSet(EntityPotionEffectEvent.Cause.UNKNOWN), EntityPotionEffectEvent.Action.CLEARED);
+                if (!event.isCancelled() && !net.neoforged.neoforge.event.EventHooks.onEffectRemoved(this, entry.getValue())) {
+                    map.put(entry.getKey(), entry.getValue());
+                }
+            }
+            map.keySet().forEach(this.activeEffects::remove);
+            this.onEffectsRemoved(map.values());
+            return true;
         }
+    }
 
-        Map<Holder<MobEffect>, MobEffectInstance> copy = Maps.newHashMap(this.activeEffects);
-        this.activeEffects.clear();
-        this.onEffectsRemoved(copy.values());
-        return true;
+    public boolean removeAllEffects(EntityPotionEffectEvent.Cause cause) {
+        removeAllEffects$cause(cause);
+        return removeAllEffects();
     }
 
     public Collection<MobEffectInstance> getActiveEffects() {
@@ -1010,30 +_,88 @@
     }
 
     public final boolean addEffect(MobEffectInstance newEffect) {
-        return this.addEffect(newEffect, null);
-    }
-
-    public boolean addEffect(MobEffectInstance newEffect, @Nullable Entity source) {
-        if (!this.canBeAffected(newEffect)) {
+        return this.addEffect(newEffect, (Entity) null);
+    }
+
+    // CraftBukkit start
+    public boolean addEffect(final MobEffectInstance newEffect, final EntityPotionEffectEvent.Cause cause) {
+        return this.addEffect(newEffect, (Entity)null, cause);
+    }
+
+    public boolean addEffect(final MobEffectInstance newEffect, final @Nullable Entity source) {
+        return this.addEffect(newEffect, source, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.UNKNOWN);
+    }
+
+    public boolean addEffect(final MobEffectInstance newEffect, final @Nullable Entity source, final EntityPotionEffectEvent.Cause cause) {
+        // Paper start - Don't fire sync event during generation
+        return this.addEffect(newEffect, source, cause, true);
+    }
+
+    public boolean addEffect(final MobEffectInstance newEffect, final @Nullable Entity source, final EntityPotionEffectEvent.Cause cause, final boolean fireEvent) {
+        // Paper end - Don't fire sync event during generation
+        if (this.tickingEffects) {
+            this.deferredEffectUpdates.offer(() -> this.addEffect(newEffect, source, cause, fireEvent));
+            return true;
+        }
+        // CraftBukkit end
+        if (!net.neoforged.neoforge.common.CommonHooks.canMobEffectBeApplied(this, newEffect, source)) {
             return false;
-        }
-
-        MobEffectInstance effect = this.activeEffects.get(newEffect.getEffect());
-        boolean changed = false;
-        if (effect == null) {
-            this.activeEffects.put(newEffect.getEffect(), newEffect);
-            this.onEffectAdded(newEffect, source);
-            changed = true;
-            newEffect.onEffectAdded(this);
-        } else if (effect.update(newEffect)) {
-            this.onEffectUpdated(effect, true, source);
-            changed = true;
-        }
-
-        newEffect.onEffectStarted(this);
-        return changed;
+        } else {
+            MobEffectInstance effect = this.activeEffects.get(newEffect.getEffect());
+            boolean changed = false;
+
+            // Paper start
+            boolean override = false;
+            boolean addAsHiddenEffect = false;
+            if (effect != null) {
+                override = new MobEffectInstance(effect).update(newEffect);
+                addAsHiddenEffect = effect.getAmplifier() > newEffect.getAmplifier() && effect.isShorterDurationThan(newEffect);
+            }
+
+            if (fireEvent) {
+                EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, effect, newEffect, source, cause, null, override);
+                override = event.isOverride();
+                if (event.isCancelled()) {
+                    return false;
+                }
+            }
+            // Paper end
+
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.living.MobEffectEvent.Added(this, effect, newEffect, source));
+            if (effect == null) {
+                this.activeEffects.put(newEffect.getEffect(), newEffect);
+                this.onEffectAdded(newEffect, source);
+                changed = true;
+                newEffect.onEffectAdded(this);
+                // CraftBukkit start
+            } else if (override) {
+                effect.update(newEffect);
+                this.onEffectUpdated(effect, true, source);
+                // CraftBukkit end
+                changed = true;
+                // Paper start - Properly update hidden effects
+            } else if (addAsHiddenEffect) {
+                if (effect.hiddenEffect == null) {
+                    effect.hiddenEffect = new MobEffectInstance(newEffect);
+                } else {
+                    effect.hiddenEffect.update(newEffect);
+                }
+                // Paper end - Properly update hidden effects
+            }
+
+            newEffect.onEffectStarted(this);
+            return changed;
+        }
     }
 
+    /**
+     * Neo: Override-Only. Call via {@link net.neoforged.neoforge.common.CommonHooks#canMobEffectBeApplied(LivingEntity, MobEffectInstance, Entity)}
+     *
+     * @param newEffect A mob effect instance
+     * @return If the mob effect instance can be applied to this entity
+     */
+    @Deprecated
+    @org.jetbrains.annotations.ApiStatus.OverrideOnly
     public boolean canBeAffected(MobEffectInstance newEffect) {
         if (this.is(EntityTypeTags.IMMUNE_TO_INFESTED)) {
             return !newEffect.is(MobEffects.INFESTED);
@@ -1045,7 +_,7 @@
     }
 
     public void forceAddEffect(MobEffectInstance newEffect, @Nullable Entity source) {
-        if (this.canBeAffected(newEffect)) {
+        if (net.neoforged.neoforge.common.CommonHooks.canMobEffectBeApplied(this, newEffect, source)) {
             MobEffectInstance previousEffect = this.activeEffects.put(newEffect.getEffect(), newEffect);
             if (previousEffect == null) {
                 this.onEffectAdded(newEffect, source);
@@ -1061,11 +_,35 @@
     }
 
     public final @Nullable MobEffectInstance removeEffectNoUpdate(Holder<MobEffect> effect) {
+        return removeEffectNoUpdate(effect, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.UNKNOWN);
+    }
+
+    public final @Nullable MobEffectInstance removeEffectNoUpdate(Holder<MobEffect> effect, EntityPotionEffectEvent.Cause cause) {
+        if (this.tickingEffects) {
+            this.deferredEffectUpdates.offer(() -> this.removeEffectNoUpdate(effect, cause));
+            return null;
+        }
+
+        MobEffectInstance activeEffect = this.activeEffects.get(effect);
+        if (activeEffect == null) {
+            return null;
+        }
+
+        EntityPotionEffectEvent event = CraftEventFactory.callEntityPotionEffectChangeEvent(this, activeEffect, null, cause);
+        if (event.isCancelled()) {
+            return null;
+        }
+
         return this.activeEffects.remove(effect);
     }
 
     public boolean removeEffect(Holder<MobEffect> effect) {
-        MobEffectInstance effectInstance = this.removeEffectNoUpdate(effect);
+        return removeEffect(effect, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.UNKNOWN);
+    }
+
+    public boolean removeEffect(Holder<MobEffect> holder, EntityPotionEffectEvent.Cause cause) {
+        if (net.neoforged.neoforge.event.EventHooks.onEffectRemoved(this, holder)) return false;
+        MobEffectInstance effectInstance = this.removeEffectNoUpdate(holder, cause);
         if (effectInstance != null) {
             this.onEffectsRemoved(List.of(effectInstance));
             return true;
@@ -1073,6 +_,7 @@
             return false;
         }
     }
+    // Youer end
 
     protected void onEffectAdded(MobEffectInstance effect, @Nullable Entity source) {
         if (!this.level().isClientSide()) {
@@ -1155,18 +_,65 @@
         }
     }
 
+    // Youer start
+    private EntityRegainHealthEvent.RegainReason heal$regainReason = EntityRegainHealthEvent.RegainReason.CUSTOM;
+    public void pushHealReason(EntityRegainHealthEvent.RegainReason regainReason) {
+        this.heal$regainReason = regainReason;
+    }
+
+    public void heal(float f, EntityRegainHealthEvent.RegainReason regainReason) {
+        this.pushHealReason(regainReason);
+        this.heal(f);
+    }
+
     public void heal(float heal) {
+        heal = net.neoforged.neoforge.event.EventHooks.onLivingHeal(this, heal);
+        if (heal <= 0) return;
         float health = this.getHealth();
         if (health > 0.0F) {
-            this.setHealth(health + heal);
+            EntityRegainHealthEvent event = new EntityRegainHealthEvent(this.getBukkitEntity(), heal, heal$regainReason);
+            // Suppress during worldgen
+            if (this.valid) {
+                this.level().getCraftServer().getPluginManager().callEvent(event);
+            }
+            if (!event.isCancelled()) {
+                this.setHealth((float) (health + event.getAmount()));
+            }
+            // CraftBukkit end
         }
     }
 
     public float getHealth() {
+        // CraftBukkit start - Use unscaled health
+        if (this instanceof ServerPlayer serverPlayer) {
+            return (float) serverPlayer.getBukkitEntity().getHealth();
+        }
+        // CraftBukkit end
         return this.entityData.get(DATA_HEALTH_ID);
     }
 
+
+    public AtomicBoolean fake$setHealth = new AtomicBoolean(false);
+    public void fake$setHealth(boolean fake) {
+        fake$setHealth.set(fake);
+    }
     public void setHealth(float health) {
+        if (fake$setHealth.getAndSet(false)) {
+            return;
+        }
+        // Paper start - Check for NaN
+        if (Float.isNaN(health)) { health = getMaxHealth(); if (this.valid) {
+            System.err.println("[NAN-HEALTH] " + getScoreboardName() + " had NaN health set");
+        } } // Paper end - Check for NaN
+        // CraftBukkit start - Handle scaled health
+        if (this instanceof ServerPlayer serverPlayer) {
+            CraftPlayer player = serverPlayer.getBukkitEntity();
+            double realHealth = Mth.clamp(health, 0.0F, player.getMaxHealth());
+            player.setRealHealth(realHealth);
+            player.updateScaledHealth(false);
+            return;
+        }
+        // CraftBukkit end
         this.entityData.set(DATA_HEALTH_ID, Mth.clamp(health, 0.0F, this.getMaxHealth()));
     }
 
@@ -1179,7 +_,9 @@
         if (this.isInvulnerableTo(level, source)) {
             return false;
         }
-
+        if (this.isRemoved() || this.dead || this.getHealth() <= 0.0F) { // CraftBukkit - Don't allow entities that got set to dead/killed elsewhere to get damaged and die
+            return false;
+        }
         if (this.isDeadOrDying()) {
             return false;
         }
@@ -1188,25 +_,47 @@
             return false;
         }
 
+        this.damageContainers.push(new net.neoforged.neoforge.common.damagesource.DamageContainer(source, damage));
+        if (net.neoforged.neoforge.common.CommonHooks.onEntityIncomingDamage(this, this.damageContainers.peek())) {
+            this.damageContainers.pop();
+            return false;
+        }
+
         if (this.isSleeping()) {
             this.stopSleeping();
         }
 
         this.noActionTime = 0;
+        damage = this.damageContainers.peek().getNewDamage(); // Neo: enforce damage container as source of truth for damage amount
         if (damage < 0.0F) {
             damage = 0.0F;
         }
 
+        // Youer start
+        bukkitEntityDamageEvent = BukkitDamageHooks.handleEntityDamage(this, source, damage); // Transmits the mod's modified damage value to the Bukkit event
+        if (bukkitEntityDamageEvent == null || bukkitEntityDamageEvent.isCancelled()) {
+            this.damageContainers.pop();
+            return false;
+        }
+        damageContainers.peek().setNewDamage((float) bukkitEntityDamageEvent.getDamage());
+        damage = (float) bukkitEntityDamageEvent.getDamage();
+        if (source.getEntity() instanceof net.minecraft.world.entity.player.Player player) {
+            player.resetAttackStrengthTicker();
+        }
+        // Youer end
+
         float originalDamage = damage;
         ItemStack itemInUse = this.getUseItem();
         float damageBlocked = this.applyItemBlocking(level, source, damage);
         damage -= damageBlocked;
         boolean blocked = damageBlocked > 0.0F;
         if (source.is(DamageTypeTags.IS_FREEZING) && this.is(EntityTypeTags.FREEZE_HURTS_EXTRA_TYPES)) {
+            damage = applyBukkitDamageModifier(EntityDamageEvent.DamageModifier.FREEZING, damage);
             damage *= 5.0F;
         }
 
         if (source.is(DamageTypeTags.DAMAGES_HELMET) && !this.getItemBySlot(EquipmentSlot.HEAD).isEmpty()) {
+            damage = applyBukkitDamageModifier(EntityDamageEvent.DamageModifier.HARD_HAT, damage);
             this.hurtHelmet(source, damage);
             damage *= 0.75F;
         }
@@ -1214,20 +_,30 @@
         if (Float.isNaN(damage) || Float.isInfinite(damage)) {
             damage = Float.MAX_VALUE;
         }
+        this.damageContainers.peek().setNewDamage(damage); // Update container with vanilla changes
 
         boolean tookFullDamage = true;
         if (this.invulnerableTime > 10.0F && !source.is(DamageTypeTags.BYPASSES_COOLDOWN)) {
             if (damage <= this.lastHurt) {
+                // Neo: Pop the container when vanilla would abort the incoming damage due to the invulnerability ticks.
+                this.damageContainers.pop();
                 return false;
             }
+
+            // If we are going to do damage despite invulnerability (due to the damage being higher than prior), we need to setup the container differently.
+            this.damageContainers.peek().setPostAttackInvulnerabilityTicks(this.invulnerableTime);
+            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.INVULNERABILITY, this.lastHurt);
 
             this.actuallyHurt(level, source, damage - this.lastHurt);
-            this.lastHurt = damage;
+            // lastHurt, in both blocks, needs to always be set to the value of `damage` since it is pre-reduction. Using a post reduction value invalidates iframes.
+            // Since we're in the "damage while invuln is active" block, we need to make sure we don't reduce the lastHurt amount.
+            this.lastHurt = java.lang.Math.max(this.lastHurt, damage);
+            this.invulnerableTime = this.damageContainers.peek().getPostAttackInvulnerabilityTicks();
             tookFullDamage = false;
         } else {
-            this.lastHurt = damage;
-            this.invulnerableTime = 20;
             this.actuallyHurt(level, source, damage);
+            this.lastHurt = damage;
+            this.invulnerableTime = this.damageContainers.peek().getPostAttackInvulnerabilityTicks();
             this.hurtDuration = 10;
             this.hurtTime = this.hurtDuration;
         }
@@ -1286,6 +_,10 @@
             CriteriaTriggers.PLAYER_HURT_ENTITY.trigger(sourcePlayer, this, source, originalDamage, damage, blocked);
         }
 
+        // Neo: Fire our post-damage event hooks. As usual, fire the extension method first, then the event version.
+        this.onDamageTaken(this.damageContainers.peek());
+        net.neoforged.neoforge.common.CommonHooks.onLivingDamagePost(this, this.damageContainers.peek());
+        this.damageContainers.pop();
         return success;
     }
 
@@ -1318,7 +_,7 @@
         }
 
         BlocksAttacks blocksAttacks = blockingWith.get(DataComponents.BLOCKS_ATTACKS);
-        if (blocksAttacks != null && !blocksAttacks.bypassedBy().map(t -> t.contains(source.typeHolder())).orElse(false)) {
+        if (blocksAttacks != null) {
             if (source.getDirectEntity() instanceof AbstractArrow abstractArrow && abstractArrow.getPierceLevel() > 0) {
                 return 0.0F;
             } else {
@@ -1334,7 +_,24 @@
                 }
 
                 float damageBlocked = blocksAttacks.resolveBlockedDamage(source, damage, angle);
-                blocksAttacks.hurtBlockingItem(this.level(), blockingWith, this, this.getUsedItemHand(), damageBlocked);
+
+                var ev = net.neoforged.neoforge.common.CommonHooks.onDamageBlock(this, this.damageContainers.peek(), damageBlocked, !blocksAttacks.bypassedBy().map(t -> t.contains(source.typeHolder())).orElse(false));
+                // Youer start
+                float bukkit = -(float) bukkitEntityDamageEvent.getDamage(EntityDamageEvent.DamageModifier.BLOCKING);
+                if (ev.getBlockedDamage() == ev.getOriginalBlockedDamage()) {
+                    if (bukkit > 0.0F) {
+                        ev.setBlocked(true);
+                        ev.setBlockedDamage(bukkit);
+                    } else {
+                        ev.setBlocked(false);
+                    }
+                }
+                // Youer end
+                if (!ev.getBlocked()) return 0.0F;
+                damageBlocked = ev.getBlockedDamage();
+                this.damageContainers.peek().setBlockedDamage(ev);
+                blocksAttacks.hurtBlockingItem(this.level(), blockingWith, this, this.getUsedItemHand(), damageBlocked, ev.shieldDamage());
+
                 if (damageBlocked > 0.0F && !source.is(DamageTypeTags.IS_PROJECTILE) && source.getDirectEntity() instanceof LivingEntity livingEntity) {
                     this.blockUsingItem(level, livingEntity, source, damage);
                 }
@@ -1365,9 +_,9 @@
         Entity sourceEntity = source.getEntity();
         if (sourceEntity instanceof Player playerSource) {
             this.setLastHurtByPlayer(playerSource, 100);
-        } else if (sourceEntity instanceof Wolf wolf && wolf.isTame()) {
-            if (wolf.getOwnerReference() != null) {
-                this.setLastHurtByPlayer(wolf.getOwnerReference().getUUID(), 100);
+        } else if (sourceEntity instanceof TamableAnimal tamableAnimal && tamableAnimal.isTame()) {
+            if (tamableAnimal.getOwnerReference() != null) {
+                this.setLastHurtByPlayer(tamableAnimal.getOwnerReference().getUUID(), 100);
             } else {
                 this.lastHurtByPlayer = null;
                 this.lastHurtByPlayerMemoryTime = 0;
@@ -1393,18 +_,35 @@
         ItemStack protectionItem = null;
         DeathProtection protection = null;
 
+        // CraftBukkit start
+        InteractionHand foundHand = null;
+        ItemStack itemStack = ItemStack.EMPTY;
         for (InteractionHand hand : InteractionHand.values()) {
-            ItemStack itemStack = this.getItemInHand(hand);
+            itemStack = this.getItemInHand(hand);
+            // CraftBukkit end
             protection = itemStack.get(DataComponents.DEATH_PROTECTION);
-            if (protection != null) {
+            if (protection != null && net.neoforged.neoforge.common.CommonHooks.onLivingUseTotem(this, killingDamage, itemStack, hand)) {
+                foundHand = hand; // CraftBukkit
                 protectionItem = itemStack.copy();
+                // itemStack.shrink(1); // CraftBukkit - moved down
+                break;
+            }
+        }
+
+        // CraftBukkit start
+        org.bukkit.inventory.EquipmentSlot handSlot = (foundHand != null) ? org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(foundHand) : null;
+        org.bukkit.event.entity.EntityResurrectEvent event = new org.bukkit.event.entity.EntityResurrectEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), handSlot);
+        event.setCancelled(protectionItem == null);
+        this.level().getCraftServer().getPluginManager().callEvent(event);
+
+        if (event.isCancelled()) {
+            protection = null;
+        } else {
+            if (!itemStack.isEmpty()) {
                 itemStack.shrink(1);
-                break;
             }
-        }
-
-        if (protectionItem != null) {
-            if (this instanceof ServerPlayer player) {
+            if (protectionItem != null && this instanceof ServerPlayer player) {
+                // CraftBukkit end
                 player.awardStat(Stats.ITEM_USED.get(protectionItem.getItem()));
                 CriteriaTriggers.USED_TOTEM.trigger(player, protectionItem);
                 protectionItem.causeUseVibration(this, GameEvent.ITEM_INTERACT_FINISH);
@@ -1450,8 +_,12 @@
         }
     }
 
+    public List<Runnable> postDeathEventTasks = new java.util.ArrayList<>(); // Paper - after death event and death (no revive as converted entity in killedEntity)
+
     public void die(DamageSource source) {
         if (!this.isRemoved() && !this.dead) {
+            // Neo: Fire the LivingDeathEvent after validating that the entity isn't already dead, but before trying to kill it.
+            if (net.neoforged.neoforge.common.CommonHooks.onLivingDeath(this, source)) return;
             Entity sourceEntity = source.getEntity();
             LivingEntity killer = this.getKillCredit();
             if (killer != null) {
@@ -1464,7 +_,7 @@
 
             this.stopUsingItem();
             if (!this.level().isClientSide() && this.hasCustomName()) {
-                LOGGER.info("Named entity {} died: {}", this, this.getCombatTracker().getDeathMessage().getString());
+                if (org.spigotmc.SpigotConfig.logNamedDeaths) LOGGER.info("Named entity {} died: {}", this, this.getCombatTracker().getDeathMessage().getString()); // Spigot
             }
 
             this.handleKillingBlow();
@@ -1491,17 +_,24 @@
         if (this.level() instanceof ServerLevel serverLevel) {
             boolean var6 = false;
             if (killer instanceof WitherBoss) {
-                if (serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+                if (net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverLevel, killer)) {
                     BlockPos pos = this.blockPosition();
                     BlockState state = Blocks.WITHER_ROSE.defaultBlockState();
                     if (this.level().getBlockState(pos).isAir() && state.canSurvive(this.level(), pos)) {
-                        this.level().setBlock(pos, state, 3);
-                        var6 = true;
+                        // CraftBukkit start - call EntityBlockFormEvent for Wither Rose
+                        var6 = org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(this.level(), pos, state, 3, this);
+                        // CraftBukkit end
                     }
                 }
 
                 if (!var6) {
                     ItemEntity itemEntity = new ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), new ItemStack(Items.WITHER_ROSE));
+                    // CraftBukkit start
+                    org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(this.getBukkitEntity(), (org.bukkit.entity.Item) itemEntity.getBukkitEntity());
+                    if (!event.callEvent()) {
+                        return;
+                    }
+                    // CraftBukkit end
                     this.level().addFreshEntity(itemEntity);
                 }
             }
@@ -1509,6 +_,7 @@
     }
 
     protected void dropAllDeathLoot(ServerLevel level, DamageSource source) {
+        this.captureDrops(new java.util.ArrayList<>());
         boolean playerKilled = this.lastHurtByPlayerMemoryTime > 0;
         if (this.shouldDropLoot(level)) {
             this.dropFromLootTable(level, source, playerKilled);
@@ -1517,18 +_,31 @@
 
         this.dropEquipment(level);
         this.dropExperience(level, source.getEntity());
+
+        Collection<ItemEntity> drops = captureDrops(null);
+        if (!net.neoforged.neoforge.common.CommonHooks.onLivingDrops(this, source, drops, lastHurtByPlayerMemoryTime > 0))
+            drops.forEach(e -> level().addFreshEntity(e));
     }
 
     protected void dropEquipment(ServerLevel level) {
     }
 
+    public int getExpReward(ServerLevel level, @Nullable Entity entity) { // CraftBukkit
+        if (!this.wasExperienceConsumed() && (this.isAlwaysExperienceDropper() || this.lastHurtByPlayerMemoryTime > 0 && this.shouldDropExperience() && (Boolean) level.getGameRules().get(GameRules.MOB_DROPS))) {
+            return this.getExperienceReward(level, entity); // CraftBukkit
+        }
+
+        return 0; // CraftBukkit
+    }
+
     protected void dropExperience(ServerLevel level, @Nullable Entity killer) {
         if (!this.wasExperienceConsumed()
             && (
                 this.isAlwaysExperienceDropper()
                     || this.lastHurtByPlayerMemoryTime > 0 && this.shouldDropExperience() && level.getGameRules().get(GameRules.MOB_DROPS)
             )) {
-            ExperienceOrb.award(level, this.position(), this.getExperienceReward(level, killer));
+            int reward = net.neoforged.neoforge.event.EventHooks.getExperienceDrop(this, this.getLastHurtByPlayer(), this.getExperienceReward(level, killer));
+            ExperienceOrb.award((ServerLevel) this.level(), this.position(), reward);
         }
     }
 
@@ -1628,6 +_,12 @@
     }
 
     public void knockback(double power, double xd, double zd, DamageSource source, float damage, boolean comesFromEffect) {
+        net.neoforged.neoforge.event.entity.living.LivingKnockBackEvent event = net.neoforged.neoforge.common.CommonHooks.onLivingKnockBack(this, (float) power, xd, zd);
+        if (event.isCanceled()) return;
+        power = event.getStrength();
+        xd = event.getRatioX();
+        zd = event.getRatioZ();
+
         power *= 1.0 - this.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE);
         if (!(power <= 0.0)) {
             this.needsSync = true;
@@ -1703,6 +_,20 @@
         return new LivingEntity.Fallsounds(SoundEvents.GENERIC_SMALL_FALL, SoundEvents.GENERIC_BIG_FALL);
     }
 
+    // CraftBukkit start - Add delegate methods
+    public SoundEvent getHurtSound0(DamageSource damagesource) {
+        return getHurtSound(damagesource);
+    }
+
+    public SoundEvent getDeathSound0() {
+        return getDeathSound();
+    }
+
+    public SoundEvent getFallDamageSound0(int fallHeight) {
+        return getFallDamageSound(fallHeight);
+    }
+    // CraftBukkit end
+
     public Optional<BlockPos> getLastClimbablePos() {
         return this.lastClimbablePos;
     }
@@ -1713,17 +_,11 @@
         } else {
             BlockPos ladderCheckPos = this.blockPosition();
             BlockState state = this.getInBlockState();
-            if (this.isFallFlying() && state.is(BlockTags.CAN_GLIDE_THROUGH)) {
-                return false;
-            } else if (state.is(BlockTags.CLIMBABLE)) {
-                this.lastClimbablePos = Optional.of(ladderCheckPos);
-                return true;
-            } else if (state.getBlock() instanceof TrapDoorBlock && this.trapdoorUsableAsLadder(ladderCheckPos, state)) {
-                this.lastClimbablePos = Optional.of(ladderCheckPos);
-                return true;
-            } else {
-                return false;
-            }
+            if (this.isFallFlying() && state.is(BlockTags.CAN_GLIDE_THROUGH))
+                return false;
+            Optional<BlockPos> ladderPos = net.neoforged.neoforge.common.CommonHooks.isLivingOnLadder(state, level(), ladderCheckPos, this);
+            if (ladderPos.isPresent()) this.lastClimbablePos = ladderPos;
+            return ladderPos.isPresent();
         }
     }
 
@@ -1784,6 +_,11 @@
             effectiveFallDistance = fallDistance;
         }
 
+        var event = net.neoforged.neoforge.common.CommonHooks.onLivingFall(this, effectiveFallDistance, damageModifier);
+        if (event.isCanceled()) return false;
+        effectiveFallDistance = event.getDistance();
+        damageModifier = event.getDamageMultiplier();
+
         boolean damaged = super.causeFallDamage(effectiveFallDistance, damageModifier, damageSource);
         int dmg = this.calculateFallDamage(effectiveFallDistance, damageModifier);
         if (dmg > 0) {
@@ -1847,10 +_,10 @@
             int xx = Mth.floor(this.getX());
             int yy = Mth.floor(this.getY() - 0.2F);
             int zz = Mth.floor(this.getZ());
-            BlockState state = this.level().getBlockState(new BlockPos(xx, yy, zz));
+            var pos = new BlockPos(xx, yy, zz);
+            BlockState state = this.level().getBlockState(pos);
             if (!state.isAir()) {
-                SoundType soundType = state.getSoundType();
-                this.playSound(soundType.getFallSound(), soundType.getVolume() * 0.5F, soundType.getPitch() * 0.75F);
+                state.playFallSound(this.level(), pos, this);
             }
         }
     }
@@ -1875,6 +_,8 @@
         if (!(damage <= 0.0F)) {
             int durabilityDamage = (int)Math.max(1.0F, damage / 4.0F);
 
+            net.neoforged.neoforge.common.CommonHooks.onArmorHurt(damageSource, slots, durabilityDamage, this);
+            if (true) return; //Neo: Invalidates the loop. Armor damage happens in common hook.
             for (EquipmentSlot slot : slots) {
                 ItemStack itemStack = this.getItemBySlot(slot);
                 Equippable equippable = itemStack.get(DataComponents.EQUIPPABLE);
@@ -1885,9 +_,10 @@
         }
     }
 
-    protected float getDamageAfterArmorAbsorb(DamageSource damageSource, float damage) {
+    public AtomicBoolean calDamageEvent = new AtomicBoolean(false);
+    public float getDamageAfterArmorAbsorb(DamageSource damageSource, float damage) {
         if (!damageSource.is(DamageTypeTags.BYPASSES_ARMOR)) {
-            this.hurtArmor(damageSource, damage);
+            if (!calDamageEvent.getAndSet(false)) this.hurtArmor(damageSource, damage);
             damage = CombatRules.getDamageAfterAbsorb(
                 this, damage, damageSource, this.getArmorValue(), (float)this.getAttributeValue(Attributes.ARMOR_TOUGHNESS)
             );
@@ -1896,7 +_,19 @@
         return damage;
     }
 
-    protected float getDamageAfterMagicAbsorb(DamageSource damageSource, float damage) {
+    private float applyBukkitDamageModifier(EntityDamageEvent.DamageModifier modifier, float currentDamage) {
+        if (bukkitEntityDamageEvent != null && bukkitEntityDamageEvent.isApplicable(modifier)) {
+            double baseDamage = bukkitEntityDamageEvent.getDamage();
+            if (bukkitEntityDamageEvent.isStillOriginal(modifier, baseDamage, currentDamage)) {
+                return (float) (baseDamage + bukkitEntityDamageEvent.getDamage(modifier));
+            } else {
+                return (float) (baseDamage + (currentDamage - baseDamage));
+            }
+        }
+        return currentDamage;
+    }
+
+    public float getDamageAfterMagicAbsorb(DamageSource damageSource, float damage) {
         if (damageSource.is(DamageTypeTags.BYPASSES_EFFECTS)) {
             return damage;
         }
@@ -1907,8 +_,10 @@
             float v = damage * absorb;
             float oldDamage = damage;
             damage = Math.max(v / 25.0F, 0.0F);
+            damage = applyBukkitDamageModifier(EntityDamageEvent.DamageModifier.RESISTANCE, damage);
             float damageResisted = oldDamage - damage;
             if (damageResisted > 0.0F && damageResisted < 3.4028235E37F) {
+                this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.MOB_EFFECTS, damageResisted);
                 if (this instanceof ServerPlayer serverPlayer) {
                     serverPlayer.awardStat(Stats.DAMAGE_RESISTED, Math.round(damageResisted * 10.0F));
                 } else if (damageSource.getEntity() instanceof ServerPlayer) {
@@ -1934,6 +_,7 @@
 
         if (enchantmentArmor > 0.0F) {
             damage = CombatRules.getDamageAfterMagicAbsorb(damage, enchantmentArmor);
+            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ENCHANTMENTS,this.damageContainers.peek().getNewDamage() - damage);
         }
 
         return damage;
@@ -1941,22 +_,41 @@
 
     protected void actuallyHurt(ServerLevel level, DamageSource source, float dmg) {
         if (!this.isInvulnerableTo(level, source)) {
-            dmg = this.getDamageAfterArmorAbsorb(source, dmg);
-            dmg = this.getDamageAfterMagicAbsorb(source, dmg);
-            float originalDamage = dmg;
-            dmg = Math.max(dmg - this.getAbsorptionAmount(), 0.0F);
-            this.setAbsorptionAmount(this.getAbsorptionAmount() - (originalDamage - dmg));
-            float absorbedDamage = originalDamage - dmg;
+            // Youer start
+            float armorAbsorb = this.getDamageAfterArmorAbsorb(source, this.damageContainers.peek().getNewDamage());
+            armorAbsorb = applyBukkitDamageModifier(EntityDamageEvent.DamageModifier.ARMOR, armorAbsorb);
+            // Youer end
+            // Neo (ImplNote): Keep this in-sync with the equivalent patches in Player#actuallyHurt.
+            // Neo: Primary damage pipeline hooks. Compute the armor and enchantment reductions against the "original" damage, and then fire LivingDamageEvent.Pre
+            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ARMOR, this.damageContainers.peek().getNewDamage() - armorAbsorb);
+            this.getDamageAfterMagicAbsorb(source, this.damageContainers.peek().getNewDamage());
+
+            // LivingDamageEvent.Pre is then able to mutate the container to change the reductions and change the "real" damage that will be inflicted.
+            float damage = net.neoforged.neoforge.common.CommonHooks.onLivingDamagePre(this, this.damageContainers.peek());
+            // Throw an exception if someone kills the entity during Pre. Vanilla is going to kill any entity which isDeadOrDying() when it bubbles back up to hurtServer, so it dying early is invalid.
+            com.google.common.base.Preconditions.checkArgument(!this.dead, "It is invalid to have killed an entity during LivingDamageEvent.Pre. Use LivingIncomingDamageEvent or LivingDamageEvent.Post instead.");
+
+            // Store a copy of the "true" inflicted damage so we can propagate it back to hurtServer.
+            this.damageContainers.peek().captureInflictedDamage();
+
+            // Next we attribute damage to absorption hearts. We record the amount blocked by absorption in the container for later calculations.
+            // Mods reduction functions may adjust the split of damage between absorption and health, so we need to first apply the reductions and then compute the amounts that go to each pool.
+            this.damageContainers.peek().setReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION, Math.min(this.getAbsorptionAmount(), damage));
+            float absorbedDamage = Math.min(damage, this.damageContainers.peek().getReduction(net.neoforged.neoforge.common.damagesource.DamageContainer.Reduction.ABSORPTION));
+            this.setAbsorptionAmount(Math.max(0, this.getAbsorptionAmount() - absorbedDamage));
+            dmg = this.damageContainers.peek().getNewDamage();
+
             if (absorbedDamage > 0.0F && absorbedDamage < 3.4028235E37F && source.getEntity() instanceof ServerPlayer serverPlayer) {
                 serverPlayer.awardStat(Stats.DAMAGE_DEALT_ABSORBED, Math.round(absorbedDamage * 10.0F));
             }
 
+            // Last thing here, inflict damage to the entity's health, based on how much damage is remaining after absorption has been processes.
             if (dmg != 0.0F) {
                 this.getCombatTracker().recordDamage(source, dmg);
                 this.setHealth(this.getHealth() - dmg);
-                this.setAbsorptionAmount(this.getAbsorptionAmount() - dmg);
                 this.gameEvent(GameEvent.ENTITY_DAMAGE);
             }
+            // Neo: Now we go back to hurtServer, where we'll update various metadata fields, process the side effects of damage (i.e. killing the entity), and firing LivingDamageEvent.Post
         }
     }
 
@@ -1985,8 +_,18 @@
     }
 
     public final void setArrowCount(int count) {
-        this.entityData.set(DATA_ARROW_COUNT_ID, count);
-    }
+        // CraftBukkit start
+        setArrowCount(count, false);
+    }
+
+    public final void setArrowCount(int i, boolean flag) {
+        ArrowBodyCountChangeEvent event = CraftEventFactory.callArrowBodyCountChangeEvent(this, getArrowCount(), i, flag);
+        if (event.isCancelled()) {
+            return;
+        }
+        this.entityData.set(LivingEntity.DATA_ARROW_COUNT_ID, event.getNewAmount());
+    }
+    // CraftBukkit end
 
     public final int getStingerCount() {
         return this.entityData.get(DATA_STINGER_COUNT_ID);
@@ -2014,6 +_,8 @@
     }
 
     public void swing(InteractionHand hand, boolean sendToSwingingEntity) {
+        ItemStack stack = this.getItemInHand(hand);
+        if (!stack.isEmpty() && stack.onEntitySwing(this, hand)) return;
         if (!this.swinging || this.swingTime >= this.getCurrentSwingDuration() / 2 || this.swingTime < 0) {
             this.swingTime = -1;
             this.swinging = true;
@@ -2156,8 +_,10 @@
 
     private void swapHandItems() {
         ItemStack tmp = this.getItemBySlot(EquipmentSlot.OFFHAND);
-        this.setItemSlot(EquipmentSlot.OFFHAND, this.getItemBySlot(EquipmentSlot.MAINHAND));
-        this.setItemSlot(EquipmentSlot.MAINHAND, tmp);
+        var event = net.neoforged.neoforge.common.CommonHooks.onLivingSwapHandItems(this);
+        if (event.isCanceled()) return;
+        this.setItemSlot(EquipmentSlot.OFFHAND, event.getItemSwappedToOffHand());
+        this.setItemSlot(EquipmentSlot.MAINHAND, event.getItemSwappedToMainHand());
     }
 
     @Override
@@ -2274,7 +_,13 @@
     }
 
     public void setItemSlot(EquipmentSlot slot, ItemStack itemStack) {
-        this.onEquipItem(slot, this.equipment.set(slot, itemStack), itemStack);
+        setItemSlot(slot, itemStack, false);
+    }
+    /** Neo: Passing {@code true} for {@code insideTransaction} sets the item without side-effects (calling {@link #onEquipItem}).
+     * Callers are responsible for performing the deferred side-effects, when the active transaction commits. */
+    public void setItemSlot(EquipmentSlot slot, ItemStack itemStack, boolean insideTransaction) {
+        var oldStack = this.equipment.set(slot, itemStack);
+        if (!insideTransaction) this.onEquipItem(slot, oldStack, itemStack);
     }
 
     public float getArmorCoverPercentage() {
@@ -2379,15 +_,18 @@
             }
 
             this.needsSync = true;
+            net.neoforged.neoforge.common.CommonHooks.onLivingJump(this);
         }
     }
 
+    @Deprecated // FORGE: use sinkInFluid instead
     protected void goDownInWater() {
-        this.setDeltaMovement(this.getDeltaMovement().add(0.0, -0.04F, 0.0));
+        this.sinkInFluid(net.neoforged.neoforge.common.NeoForgeMod.WATER_TYPE.value());
     }
 
+    @Deprecated // FORGE: use jumpInFluid instead
     protected void jumpInLiquid(TagKey<Fluid> type) {
-        this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.04F, 0.0));
+        this.setDeltaMovement(this.getDeltaMovement().add(0.0D, (double)0.04F * this.getAttributeValue(net.neoforged.neoforge.common.NeoForgeMod.SWIM_SPEED), 0.0D));
     }
 
     protected float getWaterSlowDown() {
@@ -2451,7 +_,7 @@
         BlockPos posBelow = this.getBlockPosBelowThatAffectsMyMovement();
         float blockFriction = this.onGround()
             ? computeModifiedFriction(
-                this.level().getBlockState(posBelow).getBlock().getFriction(), (float)this.getAttributeValue(Attributes.FRICTION_MODIFIER)
+                this.level().getBlockState(posBelow).getFriction(this.level(), posBelow, this), (float)this.getAttributeValue(Attributes.FRICTION_MODIFIER)
             )
             : 1.0F;
         Vec3 movement = this.handleRelativeFrictionAndCalculateMovement(input, blockFriction);
@@ -2483,7 +_,15 @@
         return computeModifiedFriction(this.omnidirectionalAirMover() ? 0.91F : 0.98F, (float)this.getAttributeValue(Attributes.AIR_DRAG_MODIFIER));
     }
 
+    /**
+     * @deprecated Neo: use {@link #travelInFluid(Vec3, FluidState)} instead
+     */
+    @Deprecated
     protected void travelInFluid(Vec3 input) {
+        travelInFluid(input, net.minecraft.world.level.material.Fluids.EMPTY.defaultFluidState());
+    }
+
+    private void travelInFluid(Vec3 input, FluidState fluidState) {
         boolean isFalling = this.getDeltaMovement().y <= 0.0;
         double oldY = this.getY();
         double baseGravity = this.getEffectiveGravity();
@@ -2512,6 +_,7 @@
             slowDown = 0.96F;
         }
 
+        speed *= (float)this.getAttributeValue(net.neoforged.neoforge.common.NeoForgeMod.SWIM_SPEED);
         this.moveRelative(speed, input);
         this.move(MoverType.SELF, this.getDeltaMovement());
         Vec3 movement = this.getDeltaMovement();
@@ -2577,6 +_,11 @@
     }
 
     public void stopFallFlying() {
+        // CraftBukkit start
+        if (org.bukkit.craftbukkit.event.CraftEventFactory.callToggleGlideEvent(this, false).isCancelled()) {
+            return;
+        }
+        // CraftBukkit end
         this.setSharedFlag(7, true);
         this.setSharedFlag(7, false);
     }
@@ -2689,7 +_,7 @@
             double xd = Mth.clamp(delta.x, -0.15F, 0.15F);
             double zd = Mth.clamp(delta.z, -0.15F, 0.15F);
             double yd = Math.max(delta.y, -0.15F);
-            if (yd < 0.0 && !this.getInBlockState().is(Blocks.SCAFFOLDING) && this.isSuppressingSlidingDownLadder() && this instanceof Player) {
+            if (yd < 0.0D && !this.getInBlockState().isScaffolding(this) && this.isSuppressingSlidingDownLadder() && this instanceof Player) {
                 yd = 0.0;
             }
 
@@ -2917,6 +_,11 @@
     }
 
     public void onAttack() {
+        // Paper start
+        this.onAttack(null);
+    }
+    public void onAttack(Entity entity) {
+        // Paper end
     }
 
     public void detectEquipmentUpdates() {
@@ -2936,6 +_,7 @@
             ItemStack previous = lastEquipmentItems.get(slot);
             ItemStack current = this.getItemBySlot(slot);
             if (this.equipmentHasChanged(previous, current)) {
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.living.LivingEquipmentChangeEvent(this, slot, previous, current));
                 if (changedItems == null) {
                     changedItems = Maps.newEnumMap(EquipmentSlot.class);
                 }
@@ -3086,10 +_,10 @@
                         this.noJumpDelay = 10;
                     }
                 } else {
-                    this.jumpInLiquid(FluidTags.LAVA);
+                    this.jumpInFluid(net.neoforged.neoforge.common.NeoForgeMod.LAVA_TYPE.value());
                 }
             } else {
-                this.jumpInLiquid(FluidTags.WATER);
+                this.jumpInFluid(net.neoforged.neoforge.common.NeoForgeMod.WATER_TYPE.value());
             }
         } else {
             this.noJumpDelay = 0;
@@ -3167,6 +_,7 @@
         this.checkFallDistanceAccumulation();
         if (!this.level().isClientSide()) {
             if (!this.canGlide()) {
+                if (this.getSharedFlag(7) != false && !CraftEventFactory.callToggleGlideEvent(this, false).isCancelled()) // CraftBukkit
                 this.setSharedFlag(7, false);
                 return;
             }
@@ -3340,13 +_,20 @@
 
     @Override
     public boolean isPickable() {
-        return !this.isRemoved();
+        return !this.isRemoved() && this.collides; // CraftBukkit
     }
 
     @Override
     public boolean isPushable() {
-        return this.isAlive() && !this.isSpectator() && !this.onClimbable();
-    }
+        return this.isAlive() && !this.isSpectator() && !this.onClimbable() && this.collides; // CraftBukkit
+    }
+
+    // CraftBukkit start - collidable API
+    @Override
+    public boolean canCollideWithBukkit(Entity entity) {
+        return isPushable() && this.collides != this.collidableExemptions.contains(entity.getUUID());
+    }
+    // CraftBukkit end
 
     @Override
     public float getYHeadRot() {
@@ -3404,10 +_,22 @@
         return (this.entityData.get(DATA_LIVING_ENTITY_FLAGS) & 2) > 0 ? InteractionHand.OFF_HAND : InteractionHand.MAIN_HAND;
     }
 
+    // Paper start - Properly cancel usable items
+    public void resyncUsingItem(ServerPlayer serverPlayer) {
+        this.resendPossiblyDesyncedDataValues(java.util.List.of(DATA_LIVING_ENTITY_FLAGS), serverPlayer);
+    }
+    // Paper end - Properly cancel usable items
+    // Paper start - lag compensate eating
+    protected long eatStartTime;
+    protected int totalEatTimeTicks;
+    // Paper end - lag compensate eating
     private void updatingUsingItem() {
         if (this.isUsingItem()) {
-            if (ItemStack.isSameItem(this.getItemInHand(this.getUsedItemHand()), this.useItem)) {
-                this.useItem = this.getItemInHand(this.getUsedItemHand());
+            ItemStack itemStack = this.getItemInHand(this.getUsedItemHand());
+            if (net.neoforged.neoforge.common.CommonHooks.canContinueUsing(this.useItem, itemStack)) {
+                this.useItem = itemStack;
+            }
+            if (itemStack == this.useItem) {
                 this.updateUsingItem(this.useItem);
             } else {
                 this.stopUsingItem();
@@ -3450,8 +_,11 @@
     }
 
     protected void updateUsingItem(ItemStack useItem) {
+        if (!useItem.isEmpty())
+            this.useItemRemaining = net.neoforged.neoforge.event.EventHooks.onItemUseTick(this, useItem, this.getUseItemRemainingTicks());
+        if (this.getUseItemRemainingTicks() > 0)
         useItem.onUseTick(this.level(), this, this.getUseItemRemainingTicks());
-        if (--this.useItemRemaining == 0 && !this.level().isClientSide() && !useItem.useOnRelease()) {
+        if (--this.useItemRemaining <= 0 && !this.level().isClientSide() && !useItem.useOnRelease()) {
             this.completeUsingItem();
         }
     }
@@ -3479,8 +_,10 @@
     public void startUsingItem(InteractionHand hand) {
         ItemStack itemStack = this.getItemInHand(hand);
         if (!itemStack.isEmpty() && !this.isUsingItem()) {
+            int duration = net.neoforged.neoforge.event.EventHooks.onItemUseStart(this, itemStack, hand, itemStack.getUseDuration(this));
+            if (duration < 0) return; // Neo: Early return for negative values, as that indicates event cancellation.
             this.useItem = itemStack;
-            this.useItemRemaining = itemStack.getUseDuration(this);
+            this.useItemRemaining = duration;
             if (!this.level().isClientSide()) {
                 this.setLivingEntityFlag(1, true);
                 this.setLivingEntityFlag(2, hand == InteractionHand.OFF_HAND);
@@ -3550,7 +_,32 @@
                 this.releaseUsingItem();
             } else {
                 if (!this.useItem.isEmpty() && this.isUsingItem()) {
-                    ItemStack result = this.useItem.finishUsingItem(this.level(), this);
+                    // CraftBukkit start - fire PlayerItemConsumeEvent
+                    ItemStack itemstack;
+                    if (this instanceof ServerPlayer entityPlayer) {
+                        org.bukkit.inventory.ItemStack craftItem = org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(this.useItem);
+                        org.bukkit.inventory.EquipmentSlot bukkitHand = org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand);
+                        org.bukkit.event.player.PlayerItemConsumeEvent event = new org.bukkit.event.player.PlayerItemConsumeEvent((org.bukkit.entity.Player) this.getBukkitEntity(), craftItem, bukkitHand);
+                        this.level().getCraftServer().getPluginManager().callEvent(event);
+
+                        if (event.isCancelled()) {
+                            // Update client
+                            Consumable consumable = this.useItem.get(DataComponents.CONSUMABLE);
+                            if (consumable != null) {
+                                consumable.cancelUsingItem(entityPlayer, this.useItem);
+                            }
+                            entityPlayer.getBukkitEntity().updateInventory();
+                            entityPlayer.getBukkitEntity().updateScaledHealth();
+                            return;
+                        }
+
+                        itemstack = (craftItem.equals(event.getItem())) ? this.useItem.finishUsingItem(this.level(), this) : org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem()).finishUsingItem(this.level(), this);
+                    } else {
+                        itemstack = this.useItem.finishUsingItem(this.level(), this);
+                    }
+                    // CraftBukkit end
+                    ItemStack copy = this.useItem.copy();
+                    ItemStack result = net.neoforged.neoforge.event.EventHooks.onItemUseFinish(this, copy, getUseItemRemainingTicks(), itemstack);
                     if (result != this.useItem) {
                         this.setItemInHand(hand, result);
                     }
@@ -3584,7 +_,11 @@
         ItemStack itemInUsedHand = this.getItemInHand(this.getUsedItemHand());
         if (!this.useItem.isEmpty() && ItemStack.isSameItem(itemInUsedHand, this.useItem)) {
             this.useItem = itemInUsedHand;
+            if (!net.neoforged.neoforge.event.EventHooks.onUseItemStop(this, useItem, this.getUseItemRemainingTicks())) {
+            ItemStack copy = this instanceof Player ? useItem.copy() : null;
             this.useItem.releaseUsing(this.level(), this, this.getUseItemRemainingTicks());
+            if (copy != null && useItem.isEmpty()) net.neoforged.neoforge.event.EventHooks.onPlayerDestroyItem((Player)this, copy, getUsedItemHand());
+            }
             if (this.useItem.useOnRelease()) {
                 this.updatingUsingItem();
             }
@@ -3594,6 +_,7 @@
     }
 
     public void stopUsingItem() {
+        if (this.isUsingItem() && !this.useItem.isEmpty()) this.useItem.onStopUsing(this, useItemRemaining);
         if (!this.level().isClientSide()) {
             boolean wasUsingItem = this.isUsingItem();
             this.recentKineticEnemies = null;
@@ -3627,6 +_,60 @@
         return null;
     }
 
+    // CraftBukkit start
+    @Override
+    public float getBukkitYaw() {
+        return this.getYHeadRot();
+    }
+    // CraftBukkit end
+
+    // Paper start
+    public HitResult getRayTrace(int maxDistance, ClipContext.Fluid fluidCollisionOption) {
+        if (maxDistance < 1 || maxDistance > 120) {
+            throw new IllegalArgumentException("maxDistance must be between 1-120");
+        }
+
+        Vec3 start = new Vec3(getX(), getY() + getEyeHeight(), getZ());
+        org.bukkit.util.Vector dir = getBukkitEntity().getLocation().getDirection().multiply(maxDistance);
+        Vec3 end = new Vec3(start.x + dir.getX(), start.y + dir.getY(), start.z + dir.getZ());
+        ClipContext raytrace = new ClipContext(start, end, ClipContext.Block.OUTLINE, fluidCollisionOption, this);
+
+        return this.level().clip(raytrace);
+    }
+
+    public net.minecraft.world.phys.@Nullable EntityHitResult getTargetEntity(int maxDistance) {
+        if (maxDistance < 1 || maxDistance > 120) {
+            throw new IllegalArgumentException("maxDistance must be between 1-120");
+        }
+
+        Vec3 start = this.getEyePosition(1.0F);
+        Vec3 direction = this.getLookAngle();
+        Vec3 end = start.add(direction.x * maxDistance, direction.y * maxDistance, direction.z * maxDistance);
+
+        List<Entity> entityList = this.level().getEntities(this, getBoundingBox().expandTowards(direction.x * maxDistance, direction.y * maxDistance, direction.z * maxDistance).inflate(1.0), EntitySelector.NO_SPECTATORS.and(Entity::isPickable));
+
+        double distance = 0.0;
+        net.minecraft.world.phys.EntityHitResult result = null;
+
+        for (Entity entity : entityList) {
+            final double inflationAmount = entity.getPickRadius();
+            AABB aabb = entity.getBoundingBox().inflate(inflationAmount, inflationAmount, inflationAmount);
+            Optional<Vec3> rayTraceResult = aabb.clip(start, end);
+
+            if (rayTraceResult.isPresent()) {
+                Vec3 rayTrace = rayTraceResult.get();
+                double distanceTo = start.distanceToSqr(rayTrace);
+                if (distanceTo < distance || distance == 0.0) {
+                    result = new net.minecraft.world.phys.EntityHitResult(entity, rayTrace);
+                    distance = distanceTo;
+                }
+            }
+        }
+
+        return result;
+    }
+    // Paper end
+
     public boolean isSuppressingSlidingDownLadder() {
         return this.isShiftKeyDown();
     }
@@ -3645,6 +_,11 @@
     }
 
     public boolean randomTeleport(double xx, double yy, double zz, boolean showParticles) {
+        return randomTeleport(xx, yy, zz, showParticles, ItemStack.EMPTY);
+    }
+
+    // Neo: Add overload with the consumed item causing the random teleport
+    public boolean randomTeleport(double xx, double yy, double zz, boolean showParticles, ItemStack consumedStack) {
         double xo = this.getX();
         double yo = this.getY();
         double zo = this.getZ();
@@ -3652,6 +_,8 @@
         boolean ok = false;
         BlockPos pos = BlockPos.containing(xx, y, zz);
         Level level = this.level();
+        // Neo: Store whether we actually did a teleport to the target, to avoid teleporting the entity back to its original position if it didn't teleport away
+        boolean teleported = false;
         if (level.hasChunkAt(pos)) {
             boolean landed = false;
 
@@ -3667,14 +_,30 @@
             }
 
             if (landed) {
+                // Neo: Fire event for either item consumption or EnderMan teleport, to get the actual final position being teleported to
+                net.neoforged.neoforge.event.entity.EntityTeleportEvent event = null;
+                if (!consumedStack.isEmpty()) {
+                    event = net.neoforged.neoforge.event.EventHooks.onItemConsumptionTeleport(this, consumedStack, xx, y, zz);
+                } else if (this instanceof net.minecraft.world.entity.monster.EnderMan) {
+                    event = net.neoforged.neoforge.event.EventHooks.onEnderTeleport(this, xx, y, zz);
+                }
+                if (event != null) {
+                    xx = event.getTargetX();
+                    y = event.getTargetY();
+                    zz = event.getTargetZ();
+                }
+                if (event == null || !event.isCanceled()) {
+                teleported = true;
                 this.teleportTo(xx, y, zz);
                 if (level.noCollision(this) && !level.containsAnyLiquid(this.getBoundingBox())) {
                     ok = true;
                 }
+                }
             }
         }
 
         if (!ok) {
+            if (teleported) // Neo: Teleport back to original only if we teleported away and need to revert
             this.teleportTo(xo, yo, zo);
             return false;
         }
@@ -3755,8 +_,8 @@
         }
 
         BlockState blockState = this.level().getBlockState(bedPosition);
-        if (blockState.getBlock() instanceof BedBlock) {
-            this.level().setBlock(bedPosition, blockState.setValue(BedBlock.OCCUPIED, true), 3);
+        if (blockState.isBed(level(), bedPosition, this)) {
+            blockState.setBedOccupied(level(), bedPosition, this, true);
         }
 
         this.setPose(Pose.SLEEPING);
@@ -3771,15 +_,17 @@
     }
 
     private boolean checkBedExists() {
-        return this.getSleepingPos().map(bedPosition -> this.level().getBlockState(bedPosition).getBlock() instanceof BedBlock).orElse(false);
+        // Neo: Overwrite the vanilla instanceof BedBlock check with isBed and fire the CanContinueSleepingEvent.
+        boolean hasBed = this.getSleepingPos().map(pos -> this.level().getBlockState(pos).isBed(this.level(), pos, this)).orElse(false);
+        return net.neoforged.neoforge.event.EventHooks.canEntityContinueSleeping(this, hasBed ? null : Player.BedSleepingProblem.OTHER_PROBLEM);
     }
 
     public void stopSleeping() {
         this.getSleepingPos().filter(this.level()::hasChunkAt).ifPresent(bedPosition -> {
             BlockState state = this.level().getBlockState(bedPosition);
-            if (state.getBlock() instanceof BedBlock) {
+            if (state.isBed(level(), bedPosition, this)) {
                 Direction facing = state.getValue(BedBlock.FACING);
-                this.level().setBlock(bedPosition, state.setValue(BedBlock.OCCUPIED, false), 3);
+                state.setBedOccupied(level(), bedPosition, this, false);
                 Vec3 standUp = BedBlock.findStandUpPosition(this.getType(), this.level(), bedPosition, facing, this.getYRot()).orElseGet(() -> {
                     BlockPos above = bedPosition.above();
                     return new Vec3(above.getX() + 0.5, above.getY() + 0.1, above.getZ() + 0.5);
@@ -3799,7 +_,9 @@
 
     public @Nullable Direction getBedOrientation() {
         BlockPos bedPos = this.getSleepingPos().orElse(null);
-        return bedPos != null ? BedBlock.getBedOrientation(this.level(), bedPos) : null;
+        if (bedPos == null) return null;
+        BlockState state = this.level().getBlockState(bedPos);
+        return !state.isBed(level(), bedPos, this) ? null : state.getBedDirection(level(), bedPos);
     }
 
     @Override
@@ -3808,7 +_,7 @@
     }
 
     public ItemStack getProjectile(ItemStack heldWeapon) {
-        return ItemStack.EMPTY;
+        return net.neoforged.neoforge.common.CommonHooks.getProjectile(this, heldWeapon, ItemStack.EMPTY);
     }
 
     public static byte entityEventForEquipmentBreak(EquipmentSlot equipmentSlot) {
@@ -3860,6 +_,8 @@
     }
 
     public EquipmentSlot getEquipmentSlotForItem(ItemStack itemStack) {
+        final EquipmentSlot slot = itemStack.getEquipmentSlot();
+        if (slot != null) return slot; // FORGE: Allow modders to set a non-default equipment slot for a stack; e.g. a non-armor chestplate-slot item
         Equippable equippable = itemStack.get(DataComponents.EQUIPPABLE);
         return equippable != null && this.canUseSlot(equippable.slot()) ? equippable.slot() : EquipmentSlot.MAINHAND;
     }
@@ -4027,5 +_,51 @@
     }
 
     public record Fallsounds(SoundEvent small, SoundEvent big) {
+    }
+
+    public float calculateItemBlocking(DamageSource source, float damage) {
+        // CraftBukkit end
+        if (damage <= 0.0F) {
+            return 0.0F;
+        }
+        ItemStack itemstack = this.getItemBlockingWith();
+
+        if (itemstack == null) {
+            return 0.0F;
+        }
+        BlocksAttacks blocksattacks = (BlocksAttacks) itemstack.get(DataComponents.BLOCKS_ATTACKS);
+
+        if (blocksattacks != null && !(Boolean) blocksattacks.bypassedBy().map((holderset) -> {
+            return holderset.contains(source.typeHolder());
+        }).orElse(false)) {
+            Entity entity = source.getDirectEntity();
+
+            if (entity instanceof AbstractArrow) {
+                AbstractArrow abstractarrow = (AbstractArrow) entity;
+
+                if (abstractarrow.getPierceLevel() > 0) {
+                    return 0.0F;
+                }
+            }
+
+            Vec3 vec3 = source.getSourcePosition();
+            double d0;
+
+            if (vec3 != null) {
+                Vec3 vec31 = this.calculateViewVector(0.0F, this.getYHeadRot());
+                Vec3 vec32 = vec3.subtract(this.position());
+
+                vec32 = (new Vec3(vec32.x, 0.0D, vec32.z)).normalize();
+                d0 = Math.acos(vec32.dot(vec31));
+            } else {
+                d0 = (double) (float) Math.PI;
+            }
+
+            float f1 = blocksattacks.resolveBlockedDamage(source, damage, d0);
+
+            // CraftBukkit start
+            return f1;
+        }
+        return 0.0F;
     }
 }
