--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -4,12 +_,14 @@
 import com.google.common.collect.Maps;
 import it.unimi.dsi.fastutil.objects.Object2IntMap.Entry;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Predicate;
 import net.minecraft.advancements.triggers.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -94,6 +_,16 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.ticks.ContainerSingleItem;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.EntityUnleashEvent;
+import org.bukkit.inventory.InventoryHolder;
 import org.jspecify.annotations.Nullable;
 
 public abstract class Mob extends LivingEntity implements Targeting, EquipmentUser, Leashable {
@@ -144,6 +_,14 @@
     private Leashable.@Nullable LeashData leashData;
     private BlockPos homePosition = BlockPos.ZERO;
     private int homeRadius = -1;
+    @Nullable
+    private EntitySpawnReason spawnType;
+    private boolean spawnCancelled = false;
+    /** Neo: Prevent immediate spawning from conversions to capture conversion results for events */
+    protected boolean preventConversionSpawns = false;
+
+    public boolean aware = true; // CraftBukkit
+    public net.kyori.adventure.util.TriState despawnInPeacefulOverride = net.kyori.adventure.util.TriState.NOT_SET; // Paper - allow changing despawnInPeaceful
 
     protected Mob(EntityType<? extends Mob> type, Level level) {
         super(type, level);
@@ -248,8 +_,26 @@
         return this.asValidTarget(this.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).orElse(null));
     }
 
+    private EntityTargetEvent.TargetReason reason = EntityTargetEvent.TargetReason.UNKNOWN;
+    public AtomicBoolean canSetTarget = new AtomicBoolean(true);
+    public void setTargetReason(EntityTargetEvent.TargetReason reason) {
+        this.reason = reason;
+    }
+
     public void setTarget(@Nullable LivingEntity target) {
-        this.target = this.asValidTarget(target);
+        target = this.asValidTarget(target);
+        net.neoforged.neoforge.event.entity.living.LivingChangeTargetEvent changeTargetEvent = net.neoforged.neoforge.common.CommonHooks.onLivingChangeTarget(this, target, net.neoforged.neoforge.event.entity.living.LivingChangeTargetEvent.LivingTargetType.MOB_TARGET);
+        if(!changeTargetEvent.isCanceled()) {
+             this.target = changeTargetEvent.getNewAboutToBeSetTarget();
+        }
+    }
+
+    public boolean setTarget(LivingEntity p_21544_, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason) {
+        if (getTarget() == p_21544_) return false;
+        setTargetReason(reason);
+        setTarget(p_21544_);
+        return canSetTarget.getAndSet(true);
+        // CraftBukkit end
     }
 
     @Override
@@ -345,6 +_,12 @@
         if (!this.level().isClientSide() && this.tickCount % 5 == 0) {
             this.updateControlFlags();
         }
+
+        // Neo: Animal armor tick patch
+        if (this.canUseSlot(EquipmentSlot.BODY)) {
+            ItemStack stack = this.getBodyArmorItem();
+            if (stack.has(DataComponents.EQUIPPABLE)) stack.onAnimalArmorTick(level(), this);
+        }
     }
 
     protected void updateControlFlags() {
@@ -364,6 +_,12 @@
         return null;
     }
 
+    // CraftBukkit start - Add delegate method
+    public SoundEvent getAmbientSound0() {
+        return getAmbientSound();
+    }
+    // CraftBukkit end
+
     @Override
     protected void addAdditionalSaveData(ValueOutput output) {
         super.addAdditionalSaveData(output);
@@ -388,13 +_,26 @@
         if (this.isNoAi()) {
             output.putBoolean("NoAI", this.isNoAi());
         }
+        output.putBoolean("Bukkit.Aware", this.aware); // CraftBukkit
+        if (this.spawnType != null) {
+            output.putString("neoforge:spawn_type", this.spawnType.name());
+        }
     }
 
     @Override
     protected void readAdditionalSaveData(ValueInput input) {
         super.readAdditionalSaveData(input);
-        this.setCanPickUpLoot(input.getBooleanOr("CanPickUpLoot", false));
-        this.persistenceRequired = input.getBooleanOr("PersistenceRequired", false);
+        // CraftBukkit start - If looting or persistence is false only use it if it was set after we started using it
+        boolean canPickUpLoot = input.getBooleanOr("CanPickUpLoot", false);
+        if (isLevelAtLeast(input, 1) || canPickUpLoot) {
+            this.setCanPickUpLoot(canPickUpLoot);
+        }
+
+        boolean persistenceRequired = input.getBooleanOr("PersistenceRequired", false);
+        if (isLevelAtLeast(input, 1) || persistenceRequired) {
+            this.persistenceRequired = persistenceRequired;
+        }
+        // CraftBukkit end
         this.dropChances = input.read("drop_chances", DropChances.CODEC).orElse(DropChances.DEFAULT);
         this.readLeashData(input);
         this.homeRadius = input.getIntOr("home_radius", -1);
@@ -406,6 +_,22 @@
         this.lootTable = input.read("DeathLootTable", LootTable.KEY_CODEC);
         this.lootTableSeed = input.getLongOr("DeathLootTableSeed", 0L);
         this.setNoAi(input.getBooleanOr("NoAI", false));
+        // CraftBukkit start
+        this.aware = input.getBooleanOr("Bukkit.Aware", this.aware);
+        // CraftBukkit end
+
+        input.getString("neoforge:spawn_type")
+                .ifPresent(spawnType -> {
+                    try {
+                        this.spawnType = EntitySpawnReason.valueOf(spawnType);
+                    } catch (Exception ignored) {
+                    }
+                });
+    }
+
+    public static net.kyori.adventure.util.TriState readDespawnInPeacefulOverride(ValueInput input) {
+        return input.read("Paper.DespawnInPeacefulOverride", io.papermc.paper.util.PaperCodecs.TRI_STATE_CODEC).orElse(net.kyori.adventure.util.TriState.NOT_SET);
+        // Paper end - allow changing despawnInPeaceful
     }
 
     @Override
@@ -464,7 +_,7 @@
             && this.canPickUpLoot()
             && this.isAlive()
             && !this.dead
-            && serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
+            && net.neoforged.neoforge.event.EventHooks.canEntityGrief(serverLevel, this)) {
             Vec3i pickupReach = this.getPickupReach();
 
             for (ItemEntity entity : this.level()
@@ -520,49 +_,61 @@
 
     protected void pickUpItem(ServerLevel level, ItemEntity entity) {
         ItemStack itemStack = entity.getItem();
-        ItemStack equippedWithStack = this.equipItemIfPossible(level, itemStack.copy());
+        ItemStack equippedWithStack = this.equipItemIfPossible(level, itemStack.copy(), entity); // CraftBukkit - add item
         if (!equippedWithStack.isEmpty()) {
             this.onItemPickup(entity);
             this.take(entity, equippedWithStack.getCount());
             itemStack.shrink(equippedWithStack.getCount());
             if (itemStack.isEmpty()) {
-                entity.discard();
+                entity.discard(EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
             }
         }
     }
 
     public ItemStack equipItemIfPossible(ServerLevel level, ItemStack itemStack) {
+        // CraftBukkit start - add item
+        return this.equipItemIfPossible(level, itemStack, null);
+    }
+
+    public ItemStack equipItemIfPossible(ServerLevel level, ItemStack itemStack, ItemEntity itementity) {
+        // CraftBukkit end
         EquipmentSlot slot = this.getEquipmentSlotForItem(itemStack);
         if (!this.isEquippableInSlot(itemStack, slot)) {
             return ItemStack.EMPTY;
-        }
-
-        ItemStack current = this.getItemBySlot(slot);
-        boolean canReplace = this.canReplaceCurrentItem(itemStack, current, slot);
-        if (slot.isArmor() && !canReplace) {
-            slot = EquipmentSlot.MAINHAND;
-            current = this.getItemBySlot(slot);
-            canReplace = current.isEmpty();
-        }
-
-        if (canReplace && this.canHoldItem(itemStack)) {
-            double dropChance = this.dropChances.byEquipment(slot);
-            if (!current.isEmpty() && Math.max(this.random.nextFloat() - 0.1F, 0.0F) < dropChance) {
-                this.spawnAtLocation(level, current);
-            }
-
-            ItemStack toEquip = slot.limit(itemStack);
-            this.setItemSlotAndDropWhenKilled(slot, toEquip);
-            this.persistenceRequired = true;
-            return toEquip;
         } else {
-            return ItemStack.EMPTY;
+            ItemStack current = this.getItemBySlot(slot);
+            boolean canReplace = this.canReplaceCurrentItem(itemStack, current, slot);
+            if (slot.isArmor() && !canReplace) {
+                slot = EquipmentSlot.MAINHAND;
+                current = this.getItemBySlot(slot);
+                canReplace = current.isEmpty();
+            }
+
+            // CraftBukkit start
+            boolean canPickup = canReplace && this.canHoldItem(itemStack);
+            if (itementity != null) {
+                canPickup = !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(this, itementity, 0, !canPickup).isCancelled();
+            }
+            if (canPickup) {
+                // CraftBukkit end
+                double dropChance = this.dropChances.byEquipment(slot);
+                if (!current.isEmpty() && Math.max(this.random.nextFloat() - 0.1F, 0.0F) < dropChance) {
+                    this.spawnAtLocation(level, current);
+                }
+
+                ItemStack toEquip = slot.limit(itemStack);
+                this.setItemSlotAndDropWhenKilled(slot, toEquip);
+                return toEquip;
+            } else {
+                return ItemStack.EMPTY;
+            }
         }
     }
 
     protected void setItemSlotAndDropWhenKilled(EquipmentSlot slot, ItemStack itemStack) {
         this.setItemSlot(slot, itemStack);
         this.setGuaranteedDrop(slot);
+        this.persistenceRequired = true;
     }
 
     protected boolean canShearEquipment(Player player) {
@@ -646,6 +_,8 @@
     private double getApproximateAttributeWith(ItemStack itemStack, Holder<Attribute> attribute, EquipmentSlot slot) {
         double baseValue = this.getAttributes().hasAttribute(attribute) ? this.getAttributeBaseValue(attribute) : 0.0;
         ItemAttributeModifiers attributeModifiers = itemStack.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
+        // Neo: Respect gameplay modifiers
+        attributeModifiers = itemStack.getAttributeModifiers();
         return attributeModifiers.compute(attribute, baseValue, slot);
     }
 
@@ -683,10 +_,17 @@
         return this.isPassenger() || this.isLeashed();
     }
 
+    // Paper start - allow changing despawnInPeaceful
+    public final boolean shouldDespawnInPeaceful() {
+        return this.despawnInPeacefulOverride.toBooleanOrElse(!this.getType().isTypeAllowedInPeaceful());
+    }
+    // Paper end - allow changing despawnInPeaceful
+
     @Override
     public void checkDespawn() {
+        if (net.neoforged.neoforge.event.EventHooks.checkMobDespawn(this)) return;
         if (this.level().getDifficulty() == Difficulty.PEACEFUL && !this.getType().isAllowedInPeaceful()) {
-            this.discard();
+            this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else if (!this.isPersistenceRequired() && !this.requiresCustomPersistence()) {
             Entity player = this.level().getNearestPlayer(this, -1.0);
             if (player != null) {
@@ -694,13 +_,13 @@
                 int instantDespawnDistance = this.getType().getCategory().getDespawnDistance();
                 int despawnDistanceSqr = instantDespawnDistance * instantDespawnDistance;
                 if (distSqr > despawnDistanceSqr && this.removeWhenFarAway(distSqr)) {
-                    this.discard();
+                    this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                 }
 
                 int noDespawnDistance = this.getType().getCategory().getNoDespawnDistance();
                 int noDespawnDistanceSqr = noDespawnDistance * noDespawnDistance;
                 if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && distSqr > noDespawnDistanceSqr && this.removeWhenFarAway(distSqr)) {
-                    this.discard();
+                    this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                 } else if (distSqr < noDespawnDistanceSqr) {
                     this.noActionTime = 0;
                 }
@@ -713,6 +_,7 @@
     @Override
     protected final void serverAiStep() {
         this.noActionTime++;
+        if (!this.aware) return; // CraftBukkit
         ProfilerFiller profiler = Profiler.get();
         profiler.push("sensing");
         this.sensing.tick();
@@ -883,6 +_,51 @@
             public boolean stillValid(Player player) {
                 return player.getVehicle() == Mob.this || player.isWithinEntityInteractionRange(Mob.this, 4.0);
             }
+
+            // CraftBukkit start - add fields and methods
+            public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+            private int maxStack = MAX_STACK;
+
+            @Override
+            public List<ItemStack> getContents() {
+                return Arrays.asList(this.getTheItem());
+            }
+
+            @Override
+            public void onOpen(CraftHumanEntity who) {
+                transaction.add(who);
+            }
+
+            @Override
+            public void onClose(CraftHumanEntity who) {
+                transaction.remove(who);
+            }
+
+            @Override
+            public List<HumanEntity> getViewers() {
+                return transaction;
+            }
+
+            @Override
+            public int getMaxStackSize() {
+                return maxStack;
+            }
+
+            @Override
+            public void setMaxStackSize(int size) {
+                maxStack = size;
+            }
+
+            @Override
+            public InventoryHolder getOwner() {
+                return (InventoryHolder) Mob.this.getBukkitEntity();
+            }
+
+            @Override
+            public Location getLocation() {
+                return Mob.this.getBukkitEntity().getLocation();
+            }
+            // CraftBukkit end
         };
     }
 
@@ -1078,6 +_,11 @@
         }
     }
 
+    /**
+     * @deprecated Override-Only. External callers should call via {@link net.neoforged.neoforge.event.EventHooks#finalizeMobSpawn}.
+     */
+    @Deprecated
+    @org.jetbrains.annotations.ApiStatus.OverrideOnly
     public @Nullable SpawnGroupData finalizeSpawn(
         ServerLevelAccessor level, DifficultyInstance difficulty, EntitySpawnReason spawnReason, @Nullable SpawnGroupData groupData
     ) {
@@ -1090,6 +_,7 @@
         }
 
         this.setLeftHanded(random.nextFloat() < 0.05F);
+        this.spawnType = spawnReason;
         return groupData;
     }
 
@@ -1227,30 +_,55 @@
     public <T extends Mob> @Nullable T convertTo(
         EntityType<T> entityType, ConversionParams params, EntitySpawnReason spawnReason, ConversionParams.AfterConversion<T> afterConversion
     ) {
+        // CraftBukkit start
+        return this.convertTo(entityType, params, spawnReason, afterConversion, EntityTransformEvent.TransformReason.UNKNOWN, CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public <T extends Mob> @Nullable T convertTo(EntityType<T> entityType, ConversionParams params, EntitySpawnReason spawnReason, ConversionParams.AfterConversion<T> afterConversion, EntityTransformEvent.TransformReason transformReason, CreatureSpawnEvent.SpawnReason bukkitSpawnReason) {
+        // CraftBukkit end
         if (this.isRemoved()) {
             return null;
-        }
-
-        T newMob = (T)entityType.create(this.level(), spawnReason);
-        if (newMob == null) {
-            return null;
-        }
-
-        params.type().convert(this, newMob, params);
-        afterConversion.finalizeConversion(newMob);
-        if (this.level() instanceof ServerLevel serverLevel) {
-            serverLevel.addFreshEntity(newMob);
-        }
-
-        if (params.type().shouldDiscardAfterConversion()) {
-            this.discard();
-        }
-
-        return newMob;
+        } else {
+            T newMob = (T)entityType.create(this.level(), spawnReason);
+            if (newMob == null) {
+                return null;
+            } else {
+                params.type().convert(this, newMob, params);
+                afterConversion.finalizeConversion(newMob);
+                // CraftBukkit start
+                if (transformReason == null) {
+                    // Special handling for slime split and pig lightning
+                    return newMob;
+                }
+
+                if (CraftEventFactory.callEntityTransformEvent(this, newMob, transformReason).isCancelled()) {
+                    return null;
+                }
+
+                params.type().postConvert(this, newMob, params);
+                // CraftBukkit end
+                if (!preventConversionSpawns)
+                if (this.level() instanceof ServerLevel serverLevel) {
+                    serverLevel.addFreshEntity(newMob, bukkitSpawnReason); // CraftBukkit
+                }
+
+                if (params.type().shouldDiscardAfterConversion()) {
+                    this.discard(EntityRemoveEvent.Cause.TRANSFORMATION); // CraftBukkit - add Bukkit remove cause
+                }
+
+                return newMob;
+            }
+        }
     }
 
     public <T extends Mob> @Nullable T convertTo(EntityType<T> entityType, ConversionParams params, ConversionParams.AfterConversion<T> afterConversion) {
-        return this.convertTo(entityType, params, EntitySpawnReason.CONVERSION, afterConversion);
+        // CraftBukkit start
+        return (T) this.convertTo(entityType, params, afterConversion, EntityTransformEvent.TransformReason.UNKNOWN, CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    public <T extends Mob> @Nullable T convertTo(EntityType<T> entityType, ConversionParams params, ConversionParams.AfterConversion<T> afterConversion, EntityTransformEvent.TransformReason transformReason, CreatureSpawnEvent.SpawnReason spawnReason) {
+        return (T) this.convertTo(entityType, params, EntitySpawnReason.CONVERSION, afterConversion, transformReason, spawnReason);
+        // CraftBukkit end
     }
 
     @Override
@@ -1291,6 +_,7 @@
     public boolean startRiding(Entity entity, boolean force, boolean sendEventAndTriggers) {
         boolean result = super.startRiding(entity, force, sendEventAndTriggers);
         if (result && this.isLeashed()) {
+            this.level().getCraftServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), EntityUnleashEvent.UnleashReason.UNKNOWN)); // CraftBukkit
             this.dropLeash();
         }
 
@@ -1399,14 +_,24 @@
     }
 
     @Override
+    @Deprecated // FORGE: use jumpInFluid instead
     protected void jumpInLiquid(TagKey<Fluid> type) {
+        this.jumpInLiquidInternal(() -> super.jumpInLiquid(type));
+    }
+
+    private void jumpInLiquidInternal(Runnable onSuper) {
         if (this.getNavigation().canFloat()) {
-            super.jumpInLiquid(type);
+            onSuper.run();
         } else {
             this.setDeltaMovement(this.getDeltaMovement().add(0.0, 0.3, 0.0));
         }
     }
 
+    @Override
+    public void jumpInFluid(net.neoforged.neoforge.fluids.FluidType type) {
+        this.jumpInLiquidInternal(() -> super.jumpInFluid(type));
+    }
+
     @VisibleForTesting
     public void removeFreeWill() {
         this.removeAllGoals(goal -> true);
@@ -1472,5 +_,40 @@
 
     public float chargeSpeedModifier() {
         return 1.0F;
+    }
+
+    /**
+     * Returns the type of spawn that created this mob, if applicable.
+     * If it could not be determined, this will return null.
+     * <p>
+     * This is set via {@link Mob#finalizeSpawn}, so you should not call this from within that method, instead using the parameter.
+     */
+    @Nullable
+    public final EntitySpawnReason getSpawnType() {
+        return this.spawnType;
+    }
+
+    /**
+     * This method exists so that spawns can be cancelled from the {@link net.neoforged.neoforge.event.entity.living.FinalizeSpawnEvent FinalizeSpawnEvent}
+     * without needing to hook up an additional handler for the {@link net.neoforged.neoforge.event.entity.EntityJoinLevelEvent EntityJoinLevelEvent}.
+     * @return if this mob will be blocked from spawning during {@link Level#addFreshEntity(Entity)}
+     * @apiNote Not public-facing API.
+     */
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public final boolean isSpawnCancelled() {
+        return this.spawnCancelled;
+    }
+
+    /**
+     * Marks this mob as being disallowed to spawn during {@link Level#addFreshEntity(Entity)}.<p>
+     * @throws UnsupportedOperationException if this entity has already been {@link Entity#isAddedToLevel()} added to the level.
+     * @apiNote Not public-facing API.
+     */
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public final void setSpawnCancelled(boolean cancel) {
+        if (this.isAddedToLevel()) {
+            throw new UnsupportedOperationException("Late invocations of Mob#setSpawnCancelled are not permitted.");
+        }
+        this.spawnCancelled = cancel;
     }
 }
