--- a/net/minecraft/world/entity/OminousItemSpawner.java
+++ b/net/minecraft/world/entity/OminousItemSpawner.java
@@ -77,7 +_,7 @@
                     spawnedEntity = this.spawnProjectile(level, projectileItem, item);
                 } else {
                     spawnedEntity = new ItemEntity(level, this.getX(), this.getY(), this.getZ(), item);
-                    level.addFreshEntity(spawnedEntity);
+                    level.addFreshEntity(spawnedEntity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.OMINOUS_ITEM_SPAWNER); // Paper - fixes and addition to spawn reason API
                 }
 
                 level.levelEvent(3021, this.blockPosition(), 1);
@@ -91,7 +_,7 @@
         ProjectileItem.DispenseConfig dispenseConfig = projectileItem.createDispenseConfig();
         dispenseConfig.overrideDispenseEvent().ifPresent(event -> level.levelEvent(event, this.blockPosition(), 0));
         Direction direction = Direction.DOWN;
-        Projectile projectile = Projectile.spawnProjectileUsingShoot(
+        Projectile projectile = Projectile.spawnProjectileUsingShootDelayed( // Paper - fixes and addition to spawn reason API
             projectileItem.asProjectile(level, this.position(), item, direction),
             level,
             item,
@@ -100,7 +_,7 @@
             direction.getStepZ(),
             dispenseConfig.power(),
             dispenseConfig.uncertainty()
-        );
+        ).spawn(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.OMINOUS_ITEM_SPAWNER); // Paper - fixes and addition to spawn reason API
         projectile.setOwner(this);
         return projectile;
     }
@@ -142,7 +_,7 @@
 
     @Override
     public PushReaction getPistonPushReaction() {
-        return PushReaction.IGNORE_ENTITY;
+        return PushReaction.IGNORE;
     }
 
     @Override
