--- a/net/minecraft/world/entity/OminousItemSpawner.java
+++ b/net/minecraft/world/entity/OminousItemSpawner.java
@@ -24,7 +_,7 @@
     private static final String TAG_ITEM = "item";
     private static final EntityDataAccessor<ItemStack> DATA_ITEM = SynchedEntityData.defineId(OminousItemSpawner.class, EntityDataSerializers.ITEM_STACK);
     public static final int TICKS_BEFORE_ABOUT_TO_SPAWN_SOUND = 36;
-    public long spawnItemAfterTicks;
+    private long spawnItemAfterTicks;
 
     public OminousItemSpawner(EntityType<? extends OminousItemSpawner> p_338198_, Level p_338269_) {
         super(p_338198_, p_338269_);
@@ -164,7 +_,7 @@
         return this.getEntityData().get(DATA_ITEM);
     }
 
-    public void setItem(ItemStack p_338789_) {
+    private void setItem(ItemStack p_338789_) {
         this.getEntityData().set(DATA_ITEM, p_338789_);
     }
 }
