--- a/net/minecraft/world/entity/OminousItemSpawner.java
+++ b/net/minecraft/world/entity/OminousItemSpawner.java
@@ -89,6 +_,7 @@
                 entity = new ItemEntity(level, this.getX(), this.getY(), this.getZ(), itemstack);
             }
 
+            entity.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.OMINOUS_ITEM_SPAWNER); // Paper - Fixes and additions to the SpawnReason API
             level.addFreshEntity(entity);
             level.levelEvent(3021, this.blockPosition(), 1);
             level.gameEvent(entity, GameEvent.ENTITY_PLACE, this.position());
