--- a/net/minecraft/world/entity/SpawnPlacementTypes.java
+++ b/net/minecraft/world/entity/SpawnPlacementTypes.java
@@ -28,9 +_,10 @@
                 BlockPos blockpos = p_321783_.above();
                 BlockPos blockpos1 = p_321783_.below();
                 BlockState blockstate = p_321666_.getBlockState(blockpos1);
-                return !blockstate.isValidSpawn(p_321666_, blockpos1, p_321839_)
-                    ? false
-                    : this.isValidEmptySpawnBlock(p_321666_, p_321783_, p_321839_) && this.isValidEmptySpawnBlock(p_321666_, blockpos, p_321839_);
+                if (!blockstate.isValidSpawn(p_321666_, blockpos1, p_321839_)) {
+                    return false;
+                }
+                return this.isValidEmptySpawnBlock(p_321666_, p_321783_, p_321839_) && this.isValidEmptySpawnBlock(p_321666_, blockpos, p_321839_);
             } else {
                 return false;
             }
