--- a/net/minecraft/world/entity/SteppedInterpolationTracker.java
+++ b/net/minecraft/world/entity/SteppedInterpolationTracker.java
@@ -4,7 +_,7 @@
 import java.util.List;
 import net.minecraft.world.phys.Vec3;
 
-public class SteppedInterpolationTracker extends InterpolationHandler.NoOpInterpolationHandler implements InterpolationTracker {
+public class SteppedInterpolationTracker implements InterpolationTracker {
     private final List<PositionStep> trackedSteps = new ArrayList<>();
     private int ticksSinceLastStep = -1;
     private Vec3 predictedDelta = Vec3.ZERO;
@@ -14,12 +_,6 @@
         this.entity = entity;
     }
 
-    @Override
-    public InterpolationTracker interpolationTracker() {
-        return this;
-    }
-
-    @Override
     public void applyPredictedMovement(Vec3 delta) {
         this.predictedDelta = this.predictedDelta.add(delta);
     }
