--- /dev/null
+++ b/com/mojang/renderpearl/backend/vulkan/glsl/ShaderCompileException.java
@@ -1,0 +_,11 @@
+package com.mojang.renderpearl.backend.vulkan.glsl;
+
+import net.neoforged.api.distmarker.Dist;
+import net.neoforged.api.distmarker.OnlyIn;
+
+@OnlyIn(Dist.CLIENT)
+public class ShaderCompileException extends Exception {
+    public ShaderCompileException(String message) {
+        super(message);
+    }
+}
