--- /dev/null
+++ b/com/mojang/renderpearl/backend/vulkan/glsl/SpvcUtil.java
@@ -1,0 +_,23 @@
+package com.mojang.renderpearl.backend.vulkan.glsl;
+
+import net.neoforged.api.distmarker.Dist;
+import net.neoforged.api.distmarker.OnlyIn;
+
+@OnlyIn(Dist.CLIENT)
+class SpvcUtil {
+    private SpvcUtil() {
+    }
+
+    public static String imageDimensionToString(int dimension) {
+        return switch (dimension) {
+            case 0 -> "1D";
+            case 1 -> "2D";
+            case 2 -> "3D";
+            case 3 -> "Cube";
+            case 4 -> "Rect";
+            case 5 -> "Buffer";
+            case 6 -> "SubpassData";
+            default -> "0x" + Integer.toHexString(dimension);
+        };
+    }
+}
