--- a/net/minecraft/world/entity/animal/golem/CopperGolemAi.java
+++ b/net/minecraft/world/entity/animal/golem/CopperGolemAi.java
@@ -128,6 +_,11 @@
                 }
 
                 if (ticksSinceReachingTarget == 60) {
+                    // Purpur start - copper golem can place items in barrels or shulkers option
+                    if (container instanceof net.minecraft.world.level.block.entity.ShulkerBoxBlockEntity shulkerBoxBlockEntity && shulkerBoxBlockEntity.openCount > 0) {
+                        container.stopOpen(copperGolem);
+                    }
+                    // Purpur end - copper golem can place items in barrels or shulkers option
                     if (container.getEntitiesWithContainerOpen().contains(body)) {
                         container.stopOpen(copperGolem);
                     }
