--- a/net/minecraft/world/entity/animal/golem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/golem/CopperGolem.java
@@ -39,6 +_,7 @@
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ServerLevelAccessor;
+import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.ChestBlock;
 import net.minecraft.world.level.block.CopperGolemStatueBlock;
@@ -53,11 +_,14 @@
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.entity.EntityRemoveEvent;
 import org.jspecify.annotations.Nullable;
 
 public class CopperGolem extends AbstractGolem implements ContainerUser, Shearable {
-    private static final long IGNORE_WEATHERING_TICK = -2L;
-    private static final long UNSET_WEATHERING_TICK = -1L;
+    public static final long IGNORE_WEATHERING_TICK = -2L;
+    public static final long UNSET_WEATHERING_TICK = -1L;
     private static final int WEATHERING_TICK_FROM = 504000;
     private static final int WEATHERING_TICK_TO = 552000;
     private static final int SPIN_ANIMATION_MIN_COOLDOWN = 200;
@@ -77,7 +_,7 @@
     );
     private @Nullable BlockPos openedChestPos;
     private @Nullable UUID lastLightningBoltUUID;
-    private long nextWeatheringTick = -1L;
+    public long nextWeatheringTick = -1L;
     private int idleAnimationStartTick = 0;
     private final AnimationState idleAnimationState = new AnimationState();
     private final AnimationState interactionGetItemAnimationState = new AnimationState();
@@ -85,6 +_,7 @@
     private final AnimationState interactionDropItemAnimationState = new AnimationState();
     private final AnimationState interactionDropNoItemAnimationState = new AnimationState();
     public static final EquipmentSlot EQUIPMENT_SLOT_ANTENNA = EquipmentSlot.SADDLE;
+    @Nullable private UUID summoner; // Purpur - Summoner API
 
     public CopperGolem(EntityType<? extends AbstractGolem> type, Level level) {
         super(type, level);
@@ -98,6 +_,63 @@
         this.getBrain().setMemory(MemoryModuleType.TRANSPORT_ITEMS_COOLDOWN_TICKS, this.getRandom().nextInt(60, 100));
     }
 
+    // Purpur start - Summoner API
+    @Nullable
+    public UUID getSummoner() {
+        return summoner;
+    }
+
+    public void setSummoner(@Nullable UUID summoner) {
+        this.summoner = summoner;
+    }
+    // Purpur end - Summoner API
+
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.copperGolemRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.copperGolemRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.copperGolemControllable;
+    }
+
+    @Override
+    protected void registerGoals() {
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this));
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.copperGolemMaxHealth);
+        this.getAttribute(Attributes.STEP_HEIGHT).setBaseValue(this.level().purpurConfig.copperGolemStepHeight);
+        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.level().purpurConfig.copperGolemMovementSpeed);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.copperGolemScale);
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.copperGolemTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.copperGolemAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
     public static AttributeSupplier.Builder createAttributes() {
         return Mob.createMobAttributes().add(Attributes.MOVEMENT_SPEED, 0.2F).add(Attributes.STEP_HEIGHT, 1.0).add(Attributes.MAX_HEALTH, 12.0);
     }
@@ -168,6 +_,7 @@
         super.addAdditionalSaveData(output);
         output.putLong("next_weather_age", this.nextWeatheringTick);
         output.store("weather_state", WeatheringCopper.WeatherState.CODEC, this.getWeatherState());
+        output.storeNullable("Purpur.Summoner", net.minecraft.core.UUIDUtil.CODEC, getSummoner()); // Purpur - Summoner API
     }
 
     @Override
@@ -175,6 +_,7 @@
         super.readAdditionalSaveData(input);
         this.nextWeatheringTick = input.getLongOr("next_weather_age", -1L);
         this.setWeatherState(input.read("weather_state", WeatheringCopper.WeatherState.CODEC).orElse(WeatheringCopper.WeatherState.UNAFFECTED));
+        this.setSummoner(input.read("Purpur.Summoner", net.minecraft.core.UUIDUtil.CODEC).orElse(null)); // Purpur - Summoner API
     }
 
     @Override
@@ -216,7 +_,13 @@
         Level level = this.level();
         if (itemStack.is(Items.SHEARS) && this.readyForShearing()) {
             if (level instanceof ServerLevel serverLevel) {
-                this.shear(serverLevel, SoundSource.PLAYERS, itemStack);
+                // Paper start - call PlayerShearEntityEvent
+                java.util.List<ItemStack> drops = this.generateDefaultDrops(serverLevel, itemStack);
+                org.bukkit.event.player.PlayerShearEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerShearEntityEvent(player, this, itemStack, hand, drops);
+                if (event.isCancelled()) return InteractionResult.PASS;
+                drops = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDrops());
+                this.shear(serverLevel, SoundSource.PLAYERS, itemStack, drops);
+                // Paper end - call PlayerShearEntityEvent
                 this.gameEvent(GameEvent.SHEAR, player);
                 itemStack.hurtAndBreak(1, player, hand);
             }
@@ -285,22 +_,29 @@
 
     private void turnToStatue(ServerLevel level) {
         BlockPos pos = this.blockPosition();
-        level.setBlock(
-            pos,
-            Blocks.COPPER_GOLEM_STATUE
-                .weathering()
-                .oxidized()
-                .defaultBlockState()
-                .setValue(
-                    CopperGolemStatueBlock.POSE, CopperGolemStatueBlock.Pose.values()[this.random.nextInt(0, CopperGolemStatueBlock.Pose.values().length)]
-                )
-                .setValue(CopperGolemStatueBlock.FACING, Direction.fromYRot(this.getYRot())),
-            3
-        );
+        // Paper start - call EntityChangeBlockEvent
+        // level.setBlock(
+        //     pos,
+        BlockState newState =
+                Blocks.COPPER_GOLEM_STATUE
+                        .weathering()
+                        .oxidized()
+                        .defaultBlockState()
+                        .setValue(
+                                CopperGolemStatueBlock.POSE, CopperGolemStatueBlock.Pose.values()[this.random.nextInt(0, CopperGolemStatueBlock.Pose.values().length)]
+                        )
+                        .setValue(CopperGolemStatueBlock.FACING, Direction.fromYRot(this.getYRot()));
+        //     Block.UPDATE_ALL
+        // );
+        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this, pos, newState)) {
+            return;
+        }
+        level.setBlock(pos, newState, Block.UPDATE_ALL);
+        // Paper end - call EntityChangeBlockEvent
         if (level.getBlockEntity(pos) instanceof CopperGolemStatueBlockEntity copperGolemStatueBlockEntity) {
             copperGolemStatueBlockEntity.createStatue(this);
             this.dropPreservedEquipment(level);
-            this.discard();
+            this.discard(null); // Paper - add Bukkit remove cause
             this.playSound(SoundEvents.COPPER_GOLEM_BECOME_STATUE);
             if (this.isLeashed()) {
                 if (level.getGameRules().get(GameRules.ENTITY_DROPS)) {
@@ -430,12 +_,30 @@
     }
 
     @Override
-    public void shear(ServerLevel level, SoundSource soundSource, ItemStack tool) {
+    // Paper start
+    public void shear(final ServerLevel level, final SoundSource soundSource, final ItemStack tool, final java.util.List<ItemStack> drops) {
         level.playSound(null, this, SoundEvents.COPPER_GOLEM_SHEAR, soundSource, 1.0F, 1.0F);
-        ItemStack itemStack = this.getItemBySlot(EQUIPMENT_SLOT_ANTENNA);
         this.setItemSlot(EQUIPMENT_SLOT_ANTENNA, ItemStack.EMPTY);
-        this.spawnAtLocation(level, itemStack, 1.5F);
-    }
+        for (ItemStack drop : drops) {
+            this.spawnAtLocation(level, drop, 1.5F);
+        }
+    }
+
+    @Override
+    public void shear(final ServerLevel level, final SoundSource soundSource, final ItemStack tool) {
+        this.shear(level, soundSource, tool, this.generateDefaultDrops(level, tool));
+    }
+
+    @Override
+    public java.util.List<ItemStack> generateDefaultDrops(final ServerLevel level, final ItemStack tool) {
+        if (!this.readyForShearing()) {
+            return java.util.Collections.emptyList();
+        }
+        final java.util.List<ItemStack> drops = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(1);
+        drops.add(this.getItemBySlot(EQUIPMENT_SLOT_ANTENNA).copy());
+        return drops;
+    }
+    // Paper end
 
     @Override
     public boolean readyForShearing() {
