--- a/net/minecraft/world/entity/animal/golem/CopperGolem.java
+++ b/net/minecraft/world/entity/animal/golem/CopperGolem.java
@@ -77,7 +_,7 @@
     );
     private @Nullable BlockPos openedChestPos;
     private @Nullable UUID lastLightningBoltUUID;
-    private long nextWeatheringTick = -1L;
+    public long nextWeatheringTick = -1L;
     private int idleAnimationStartTick = 0;
     private final AnimationState idleAnimationState = new AnimationState();
     private final AnimationState interactionGetItemAnimationState = new AnimationState();
