--- a/net/minecraft/world/entity/ai/memory/MemoryModuleType.java
+++ b/net/minecraft/world/entity/ai/memory/MemoryModuleType.java
@@ -73,6 +_,7 @@
     public static final MemoryModuleType<Player> TEMPTING_PLAYER = register("tempting_player");
     public static final MemoryModuleType<Integer> TEMPTATION_COOLDOWN_TICKS = register("temptation_cooldown_ticks", Codec.INT);
     public static final MemoryModuleType<Integer> GAZE_COOLDOWN_TICKS = register("gaze_cooldown_ticks", Codec.INT);
+    public static final MemoryModuleType<Boolean> IS_TEMPTED = register("is_tempted", Codec.BOOL);
     public static final MemoryModuleType<Integer> LONG_JUMP_COOLDOWN_TICKS = register("long_jump_cooling_down", Codec.INT);
     public static final MemoryModuleType<Boolean> LONG_JUMP_MID_JUMP = register("long_jump_mid_jump");
     public static final MemoryModuleType<Boolean> HAS_HUNTING_COOLDOWN = register("has_hunting_cooldown", Codec.BOOL);
@@ -150,10 +_,12 @@
     public static final MemoryModuleType<BlockPos> BREEZE_JUMP_TARGET = register("breeze_jump_target", BlockPos.CODEC);
     public static final MemoryModuleType<Unit> BREEZE_LEAVING_WATER = register("breeze_leaving_water", Unit.CODEC);
     private final Optional<Codec<ExpirableValue<U>>> codec;
+    public final Optional<Codec<U>> rawCodec; // CraftBukkit
 
     @VisibleForTesting
     public MemoryModuleType(Optional<Codec<U>> codec) {
         this.codec = codec.map(ExpirableValue::codec);
+        this.rawCodec = codec; // CraftBukkit
     }
 
     @Override
