--- a/net/minecraft/gametest/framework/GameTestHelper.java
+++ b/net/minecraft/gametest/framework/GameTestHelper.java
@@ -68,7 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
-public class GameTestHelper implements net.neoforged.neoforge.common.extensions.GameTestHelperExtension {
+public class GameTestHelper {
     public final GameTestInfo testInfo;
     private boolean finalCheckAdded;
 
