--- a/net/minecraft/gametest/framework/TestEnvironmentDefinition.java
+++ b/net/minecraft/gametest/framework/TestEnvironmentDefinition.java
@@ -41,6 +_,7 @@
         Registry.register(registry, "all_of", TestEnvironmentDefinition.AllOf.CODEC);
         Registry.register(registry, "clock_time", TestEnvironmentDefinition.ClockTime.CODEC);
         Registry.register(registry, "difficulty", TestEnvironmentDefinition.SetDifficulty.CODEC);
+        Registry.register(registry, "dimension", TestEnvironmentDefinition.Dimension.CODEC);
         Registry.register(registry, "function", TestEnvironmentDefinition.Functions.CODEC);
         Registry.register(registry, "game_rules", TestEnvironmentDefinition.SetGameRules.CODEC);
         Registry.register(registry, "timeline_attributes", TestEnvironmentDefinition.Timelines.CODEC);
@@ -113,7 +_,7 @@
 
         public Long setup(ServerLevel level) {
             MinecraftServer server = level.getServer();
-            long previous = server.clockManager().getInstance(this.clock).totalTicks();
+            long previous = server.clockManager().getTotalTicks(this.clock);
             server.clockManager().setTotalTicks(this.clock, this.time);
             return previous;
         }
@@ -129,6 +_,45 @@
         }
     }
 
+    record Dimension(TestEnvironmentDefinition.Dimension.Type type) implements TestEnvironmentDefinition<TestEnvironmentDefinition.Dimension.Type> {
+        public static final MapCodec<TestEnvironmentDefinition.Dimension> CODEC = RecordCodecBuilder.mapCodec(
+            instance -> instance.group(TestEnvironmentDefinition.Dimension.Type.CODEC.fieldOf("dimension").forGetter(TestEnvironmentDefinition.Dimension::type))
+                .apply(instance, TestEnvironmentDefinition.Dimension::new)
+        );
+
+        public TestEnvironmentDefinition.Dimension.Type setup(ServerLevel level) {
+            return this.type;
+        }
+
+        public void teardown(ServerLevel level, TestEnvironmentDefinition.Dimension.Type savedData) {
+        }
+
+        @Override
+        public MapCodec<TestEnvironmentDefinition.Dimension> codec() {
+            return CODEC;
+        }
+
+        public enum Type implements StringRepresentable {
+            OVERWORLD("overworld"),
+            NETHER("nether"),
+            END("end");
+
+            public static final Codec<TestEnvironmentDefinition.Dimension.Type> CODEC = StringRepresentable.fromEnum(
+                TestEnvironmentDefinition.Dimension.Type::values
+            );
+            private final String id;
+
+            Type(String id) {
+                this.id = id;
+            }
+
+            @Override
+            public String getSerializedName() {
+                return this.id;
+            }
+        }
+    }
+
     record Functions(Optional<Identifier> setupFunction, Optional<Identifier> teardownFunction) implements TestEnvironmentDefinition<Unit> {
         private static final Logger LOGGER = LogUtils.getLogger();
         public static final MapCodec<TestEnvironmentDefinition.Functions> CODEC = RecordCodecBuilder.mapCodec(
@@ -177,12 +_,12 @@
 
         public Difficulty setup(ServerLevel level) {
             Difficulty oldDifficulty = level.getDifficulty();
-            level.getServer().setDifficulty(this.difficulty, true);
+            level.getServer().setDifficulty(level, this.difficulty, null, true); // Paper
             return oldDifficulty;
         }
 
         public void teardown(ServerLevel level, Difficulty saveData) {
-            level.getServer().setDifficulty(saveData, true);
+            level.getServer().setDifficulty(level, saveData, null, true); // Paper
         }
 
         @Override
@@ -201,7 +_,7 @@
             GameRuleMap originalState = GameRuleMap.of();
             GameRules gameRules = level.getGameRules();
             this.gameRulesMap.keySet().forEach(rule -> setFromActive(originalState, (GameRule<?>)rule, gameRules));
-            gameRules.setAll(this.gameRulesMap, level.getServer());
+            gameRules.setAll(this.gameRulesMap, level);
             return originalState;
         }
 
@@ -210,7 +_,7 @@
         }
 
         public void teardown(ServerLevel level, GameRuleMap saveData) {
-            level.getGameRules().setAll(saveData, level.getServer());
+            level.getGameRules().setAll(saveData, level);
         }
 
         @Override
@@ -298,7 +_,7 @@
             }
 
             public void apply(ServerLevel level) {
-                level.getServer().setWeatherParameters(this.clearTime, this.rainTime, this.raining, this.thundering);
+                level.getServer().setWeatherParameters(level, this.clearTime, this.rainTime, this.raining, this.thundering); // Paper - per-level WeatherData
             }
 
             @Override
