--- a/net/minecraft/gametest/framework/TestCommand.java
+++ b/net/minecraft/gametest/framework/TestCommand.java
@@ -179,7 +_,7 @@
         }
 
         StructureGridSpawner spawner = new StructureGridSpawner(testPos, 10, true);
-        GameTestRunner runner = GameTestRunner.Builder.fromBatches(batches, source.getServer())
+        GameTestRunner runner = GameTestRunner.Builder.fromBatches(batches, level)
             .batcher(GameTestBatchFactory.fromGameTestInfo(100))
             .newStructureSpawner(spawner)
             .existingStructureSpawner(spawner)
@@ -204,7 +_,7 @@
         } else {
             FailedTestTracker.forgetFailedTests();
             source.sendSuccess(() -> Component.translatable("commands.test.run.running", infos.size()), false);
-            GameTestRunner runner = GameTestRunner.Builder.fromInfo(infos, source.getServer())
+            GameTestRunner runner = GameTestRunner.Builder.fromInfo(infos, level)
                 .newStructureSpawner(new StructureGridSpawner(testPos, testsPerRow, false))
                 .build();
             return trackAndStartRunner(source, runner);
@@ -220,10 +_,10 @@
                 structurePos -> {
                     if (finder.source().getLevel().getBlockEntity(structurePos) instanceof TestInstanceBlockEntity testBlock) {
                         Direction var13 = testBlock.getRotation().rotate(Direction.NORTH);
-                        BlockPos teleportPosition = testBlock.getBlockPos().relative(var13, 2);
+                        BlockPos telportPosition = testBlock.getBlockPos().relative(var13, 2);
                         int teleportYRot = (int)var13.getOpposite().toYRot();
                         String tpCommand = String.format(
-                            Locale.ROOT, "/tp @s %d %d %d %d 0", teleportPosition.getX(), teleportPosition.getY(), teleportPosition.getZ(), teleportYRot
+                            Locale.ROOT, "/tp @s %d %d %d %d 0", telportPosition.getX(), telportPosition.getY(), telportPosition.getZ(), teleportYRot
                         );
                         int dx = sourcePos.getX() - structurePos.getX();
                         int dz = sourcePos.getZ() - structurePos.getZ();
@@ -427,7 +_,7 @@
                             )
                     )
             );
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (org.purpurmc.purpur.PurpurConfig.registerMinecraftDebugCommands || SharedConstants.IS_RUNNING_IN_IDE) { // Purpur - register minecraft debug commands
             testCommand = testCommand.then(
                     Commands.literal("export")
                         .then(
