--- a/net/minecraft/gametest/framework/TestCommand.java
+++ b/net/minecraft/gametest/framework/TestCommand.java
@@ -44,12 +_,10 @@
 import net.minecraft.network.chat.Style;
 import net.minecraft.network.protocol.game.ClientboundGameTestHighlightPosPacket;
 import net.minecraft.resources.Identifier;
-import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.commands.InCommandFunction;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.Mth;
-import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.Rotation;
 import net.minecraft.world.level.block.entity.BlockEntityTypes;
@@ -57,7 +_,6 @@
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.phys.BlockHitResult;
 import org.apache.commons.lang3.mutable.MutableInt;
-import org.jspecify.annotations.Nullable;
 
 public class TestCommand {
     public static final int TEST_NEARBY_SEARCH_RADIUS = 15;
@@ -179,7 +_,7 @@
         }
 
         StructureGridSpawner spawner = new StructureGridSpawner(testPos, 10, true);
-        GameTestRunner runner = GameTestRunner.Builder.fromBatches(batches, source.getServer())
+        GameTestRunner runner = GameTestRunner.Builder.fromBatches(batches, level)
             .batcher(GameTestBatchFactory.fromGameTestInfo(100))
             .newStructureSpawner(spawner)
             .existingStructureSpawner(spawner)
@@ -204,7 +_,7 @@
         } else {
             FailedTestTracker.forgetFailedTests();
             source.sendSuccess(() -> Component.translatable("commands.test.run.running", infos.size()), false);
-            GameTestRunner runner = GameTestRunner.Builder.fromInfo(infos, source.getServer())
+            GameTestRunner runner = GameTestRunner.Builder.fromInfo(infos, level)
                 .newStructureSpawner(new StructureGridSpawner(testPos, testsPerRow, false))
                 .build();
             return trackAndStartRunner(source, runner);
@@ -220,10 +_,10 @@
                 structurePos -> {
                     if (finder.source().getLevel().getBlockEntity(structurePos) instanceof TestInstanceBlockEntity testBlock) {
                         Direction var13 = testBlock.getRotation().rotate(Direction.NORTH);
-                        BlockPos teleportPosition = testBlock.getBlockPos().relative(var13, 2);
+                        BlockPos telportPosition = testBlock.getBlockPos().relative(var13, 2);
                         int teleportYRot = (int)var13.getOpposite().toYRot();
                         String tpCommand = String.format(
-                            Locale.ROOT, "/tp @s %d %d %d %d 0", teleportPosition.getX(), teleportPosition.getY(), teleportPosition.getZ(), teleportYRot
+                            Locale.ROOT, "/tp @s %d %d %d %d 0", telportPosition.getX(), telportPosition.getY(), telportPosition.getZ(), teleportYRot
                         );
                         int dx = sourcePos.getX() - structurePos.getX();
                         int dz = sourcePos.getZ() - structurePos.getZ();
@@ -427,7 +_,7 @@
                             )
                     )
             );
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (org.purpurmc.purpur.PurpurConfig.registerMinecraftDebugCommands || SharedConstants.IS_RUNNING_IN_IDE) { // Purpur - register minecraft debug commands
             testCommand = testCommand.then(
                     Commands.literal("export")
                         .then(
@@ -572,91 +_,9 @@
     }
 
     private static BlockPos createTestPositionAround(CommandSourceStack source) {
-        TestCommand.Info info = playerAndTestInfo(source);
-        return new BlockPos(info.playerPos.getX(), info.surfaceY, info.playerPos.getZ() + 3);
-    }
-
-    private static TestCommand.Info playerAndTestInfo(CommandSourceStack source) {
-        ServerPlayer serverPlayer = source.getPlayer();
-        BlockPos playerPos = BlockPos.containing(serverPlayer == null ? source.getPosition() : serverPlayer.position());
-        BlockPos testPos = new BlockPos(playerPos.getX(), 384, playerPos.getZ());
-        source.getLevel().getChunk(testPos);
-        int surfaceY = source.getLevel().getHeightmapPos(Heightmap.Types.WORLD_SURFACE, testPos).getY();
-        return new TestCommand.Info(source.getPlayer(), playerPos, testPos, surfaceY);
-    }
-
-    private static void outputCoordinates(CommandSourceStack source) {
-        TestCommand.Info info = playerAndTestInfo(source);
-        ServerPlayer player = info.player;
-        if (player != null) {
-            float playerRotation = player.getYRot();
-            float playerPitch = player.getXRot();
-            ResourceKey<Level> dimensionKey = player.level().dimension();
-            String playerDimension = dimensionKey.identifier().toString();
-            String testDimension = source.getLevel().dimension().identifier().toString();
-            if (!testDimension.equals(playerDimension)) {
-                Component playerCoordinates = ComponentUtils.wrapInSquareBrackets(
-                        Component.translatable(
-                            "test.player.coordinates",
-                            info.playerPos.getX(),
-                            info.playerPos.getY(),
-                            info.playerPos.getZ(),
-                            playerDimension.substring(playerDimension.indexOf(58) + 1)
-                        )
-                    )
-                    .withStyle(
-                        s -> s.withColor(ChatFormatting.GOLD)
-                            .withClickEvent(
-                                new ClickEvent.SuggestCommand(
-                                    "/execute in "
-                                        + playerDimension
-                                        + " run tp @s "
-                                        + info.playerPos.getX()
-                                        + " "
-                                        + info.playerPos.getY()
-                                        + " "
-                                        + info.playerPos.getZ()
-                                        + " "
-                                        + Mth.floor(playerRotation)
-                                        + " "
-                                        + Mth.floor(playerPitch)
-                                )
-                            )
-                            .withHoverEvent(new HoverEvent.ShowText(Component.translatable("chat.coordinates.tooltip")))
-                    );
-                source.sendSuccess(() -> playerCoordinates, false);
-                Component testCoordinates = ComponentUtils.wrapInSquareBrackets(
-                        Component.translatable(
-                            "test.run.coordinates",
-                            info.playerPos.getX(),
-                            info.surfaceY,
-                            info.playerPos.getZ(),
-                            testDimension.substring(testDimension.indexOf(58) + 1)
-                        )
-                    )
-                    .withStyle(
-                        s -> s.withColor(ChatFormatting.YELLOW)
-                            .withClickEvent(
-                                new ClickEvent.SuggestCommand(
-                                    "/execute in "
-                                        + testDimension
-                                        + " run tp @s "
-                                        + info.playerPos.getX()
-                                        + " "
-                                        + info.surfaceY
-                                        + " "
-                                        + info.playerPos.getZ()
-                                        + " 0 0"
-                                )
-                            )
-                            .withHoverEvent(new HoverEvent.ShowText(Component.translatable("chat.coordinates.tooltip")))
-                    );
-                source.sendSuccess(() -> testCoordinates, false);
-            }
-        }
-    }
-
-    private record Info(@Nullable ServerPlayer player, BlockPos playerPos, BlockPos testPos, int surfaceY) {
+        BlockPos playerPos = BlockPos.containing(source.getPosition());
+        int surfaceY = source.getLevel().getHeightmapPos(Heightmap.Types.WORLD_SURFACE, playerPos).getY();
+        return new BlockPos(playerPos.getX(), surfaceY, playerPos.getZ() + 3);
     }
 
     private record TestBatchSummaryDisplayer(CommandSourceStack source) implements GameTestBatchListener {
@@ -703,8 +_,6 @@
                 if (this.tracker.hasFailedOptional()) {
                     this.source.sendSystemMessage(Component.translatable("commands.test.summary.optional_failed", this.tracker.getFailedOptionalCount()));
                 }
-
-                TestCommand.outputCoordinates(this.source);
             }
         }
     }
