--- a/net/minecraft/gametest/framework/TestCommand.java
+++ b/net/minecraft/gametest/framework/TestCommand.java
@@ -13,14 +_,12 @@
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import com.mojang.brigadier.suggestion.Suggestions;
 import com.mojang.brigadier.suggestion.SuggestionsBuilder;
-import it.unimi.dsi.fastutil.objects.ReferenceArraySet;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.Optional;
-import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.Function;
 import java.util.stream.Stream;
@@ -35,7 +_,6 @@
 import net.minecraft.commands.arguments.ResourceSelectorArgument;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
-import net.minecraft.core.GlobalPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.Vec3i;
 import net.minecraft.core.registries.Registries;
@@ -48,7 +_,6 @@
 import net.minecraft.network.protocol.game.ClientboundGameTestHighlightPosPacket;
 import net.minecraft.resources.Identifier;
 import net.minecraft.resources.ResourceKey;
-import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.commands.InCommandFunction;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
@@ -110,13 +_,12 @@
     private static int clear(TestFinder finder) throws CommandSyntaxException {
         stopTests();
         CommandSourceStack source = finder.source();
-        MinecraftServer server = source.getServer();
+        ServerLevel level = source.getLevel();
         List<TestInstanceBlockEntity> tests = finder.findTestPos()
-            .flatMap(globalPos -> server.getLevel(globalPos.dimension()).getBlockEntity(globalPos.pos(), BlockEntityTypes.TEST_INSTANCE_BLOCK).stream())
+            .flatMap(pos -> level.getBlockEntity(pos, BlockEntityTypes.TEST_INSTANCE_BLOCK).stream())
             .toList();
 
         for (TestInstanceBlockEntity testInstanceBlockEntity : tests) {
-            ServerLevel level = (ServerLevel)testInstanceBlockEntity.getLevel();
             StructureUtils.clearSpaceForStructure(testInstanceBlockEntity.getTestBoundingBox(), level);
             testInstanceBlockEntity.removeBarriers();
             level.destroyBlock(testInstanceBlockEntity.getBlockPos(), false);
@@ -132,13 +_,13 @@
 
     private static int export(TestFinder finder) throws CommandSyntaxException {
         CommandSourceStack source = finder.source();
+        ServerLevel level = source.getLevel();
         int count = 0;
         boolean allGood = true;
 
-        for (Iterator<GlobalPos> iterator = finder.findTestPos().iterator(); iterator.hasNext(); count++) {
-            GlobalPos pos = iterator.next();
-            ServerLevel level = source.getServer().getLevel(pos.dimension());
-            if (!(level.getBlockEntity(pos.pos()) instanceof TestInstanceBlockEntity blockEntity)) {
+        for (Iterator<BlockPos> iterator = finder.findTestPos().iterator(); iterator.hasNext(); count++) {
+            BlockPos pos = iterator.next();
+            if (!(level.getBlockEntity(pos) instanceof TestInstanceBlockEntity blockEntity)) {
                 throw TEST_INSTANCE_COULD_NOT_BE_FOUND.create();
             }
 
@@ -159,7 +_,8 @@
     private static int verify(TestFinder finder) {
         stopTests();
         CommandSourceStack source = finder.source();
-        Function<ResourceKey<Level>, BlockPos> testPos = dimension -> createTestPositionAround(source, dimension);
+        ServerLevel level = source.getLevel();
+        BlockPos testPos = createTestPositionAround(source);
         Collection<GameTestInfo> infos = Stream.concat(
                 toGameTestInfos(source, RetryOptions.noRetries(), finder), toGameTestInfo(source, RetryOptions.noRetries(), finder, 0)
             )
@@ -172,20 +_,18 @@
                 Collection<GameTestInfo> transformedInfos = new ArrayList<>();
 
                 for (int i = 0; i < 100; i++) {
-                    GameTestInfo copyInfo = new GameTestInfo(info.getTestHolder(), rotation, info.getLevel(), new RetryOptions(1, true));
+                    GameTestInfo copyInfo = new GameTestInfo(info.getTestHolder(), rotation, level, new RetryOptions(1, true));
                     copyInfo.setTestBlockPos(info.getTestBlockPos());
                     transformedInfos.add(copyInfo);
                 }
 
-                GameTestBatch batch = GameTestBatchFactory.toGameTestBatch(
-                    transformedInfos, info.getTest().batch(), rotation.ordinal(), info.getLevel().dimension()
-                );
+                GameTestBatch batch = GameTestBatchFactory.toGameTestBatch(transformedInfos, info.getTest().batch(), rotation.ordinal());
                 batches.add(batch);
             }
         }
 
         StructureGridSpawner spawner = new StructureGridSpawner(testPos, 10, true);
-        GameTestRunner runner = GameTestRunner.Builder.fromBatches(batches, source.getServer())
+        GameTestRunner runner = GameTestRunner.Builder.fromBatches(batches, level)
             .batcher(GameTestBatchFactory.fromGameTestInfo(100))
             .newStructureSpawner(spawner)
             .existingStructureSpawner(spawner)
@@ -198,6 +_,8 @@
     private static int run(TestFinder finder, RetryOptions retryOptions, int extraRotationSteps, int testsPerRow) {
         stopTests();
         CommandSourceStack source = finder.source();
+        ServerLevel level = source.getLevel();
+        BlockPos testPos = createTestPositionAround(source);
         Collection<GameTestInfo> infos = Stream.concat(
                 toGameTestInfos(source, retryOptions, finder), toGameTestInfo(source, retryOptions, finder, extraRotationSteps)
             )
@@ -208,8 +_,7 @@
         } else {
             FailedTestTracker.forgetFailedTests();
             source.sendSuccess(() -> Component.translatable("commands.test.run.running", infos.size()), false);
-            Function<ResourceKey<Level>, BlockPos> testPos = dimension -> createTestPositionAround(source, dimension);
-            GameTestRunner runner = GameTestRunner.Builder.fromInfo(infos, source.getServer())
+            GameTestRunner runner = GameTestRunner.Builder.fromInfo(infos, level)
                 .newStructureSpawner(new StructureGridSpawner(testPos, testsPerRow, false))
                 .build();
             return trackAndStartRunner(source, runner);
@@ -219,23 +_,22 @@
     private static int locate(TestFinder finder) throws CommandSyntaxException {
         finder.source().sendSystemMessage(Component.translatable("commands.test.locate.started"));
         MutableInt structuresFound = new MutableInt(0);
-        MinecraftServer server = finder.source().getServer();
         BlockPos sourcePos = BlockPos.containing(finder.source().getPosition());
         finder.findTestPos()
             .forEach(
                 structurePos -> {
-                    if (server.getLevel(structurePos.dimension()).getBlockEntity(structurePos.pos()) instanceof TestInstanceBlockEntity testBlock) {
-                        Direction var14 = testBlock.getRotation().rotate(Direction.NORTH);
-                        BlockPos teleportPosition = testBlock.getBlockPos().relative(var14, 2);
-                        int teleportYRot = (int)var14.getOpposite().toYRot();
+                    if (finder.source().getLevel().getBlockEntity(structurePos) instanceof TestInstanceBlockEntity testBlock) {
+                        Direction var13 = testBlock.getRotation().rotate(Direction.NORTH);
+                        BlockPos telportPosition = testBlock.getBlockPos().relative(var13, 2);
+                        int teleportYRot = (int)var13.getOpposite().toYRot();
                         String tpCommand = String.format(
-                            Locale.ROOT, "/tp @s %d %d %d %d 0", teleportPosition.getX(), teleportPosition.getY(), teleportPosition.getZ(), teleportYRot
+                            Locale.ROOT, "/tp @s %d %d %d %d 0", telportPosition.getX(), telportPosition.getY(), telportPosition.getZ(), teleportYRot
                         );
-                        int dx = sourcePos.getX() - structurePos.pos().getX();
-                        int dz = sourcePos.getZ() - structurePos.pos().getZ();
+                        int dx = sourcePos.getX() - structurePos.getX();
+                        int dz = sourcePos.getZ() - structurePos.getZ();
                         int distance = Mth.floor(Mth.sqrt(dx * dx + dz * dz));
                         MutableComponent coordinates = ComponentUtils.wrapInSquareBrackets(
-                                Component.translatable("chat.coordinates", structurePos.pos().getX(), structurePos.pos().getY(), structurePos.pos().getZ())
+                                Component.translatable("chat.coordinates", structurePos.getX(), structurePos.getY(), structurePos.getZ())
                             )
                             .withStyle(
                                 s -> s.withColor(ChatFormatting.GREEN)
@@ -433,7 +_,7 @@
                             )
                     )
             );
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (org.purpurmc.purpur.PurpurConfig.registerMinecraftDebugCommands || SharedConstants.IS_RUNNING_IN_IDE) { // Purpur - register minecraft debug commands
             testCommand = testCommand.then(
                     Commands.literal("export")
                         .then(
@@ -471,27 +_,16 @@
     private static Stream<GameTestInfo> toGameTestInfo(CommandSourceStack source, RetryOptions retryOptions, TestInstanceFinder finder, int rotationSteps) {
         return finder.findTests()
             .filter(test -> verifyStructureExists(source, test.value().structure()))
-            .flatMap(
-                test -> {
-                    ResourceKey<Level> dimension = test.value().info().dimension();
-                    ServerLevel level = source.getServer().getLevel(dimension);
-                    if (level == null) {
-                        source.sendFailure(Component.literal("Could not resolve level for dimension: " + dimension.identifier()));
-                        return Stream.empty();
-                    } else {
-                        return Stream.of(
-                            new GameTestInfo(
-                                (Holder.Reference<GameTestInstance>)test, StructureUtils.getRotationForRotationSteps(rotationSteps), level, retryOptions
-                            )
-                        );
-                    }
-                }
+            .map(
+                test -> new GameTestInfo(
+                    (Holder.Reference<GameTestInstance>)test, StructureUtils.getRotationForRotationSteps(rotationSteps), source.getLevel(), retryOptions
+                )
             );
     }
 
-    private static Optional<GameTestInfo> createGameTestInfo(GlobalPos testBlockPos, CommandSourceStack source, RetryOptions retryOptions) {
-        ServerLevel level = source.getServer().getLevel(testBlockPos.dimension());
-        if (level.getBlockEntity(testBlockPos.pos()) instanceof TestInstanceBlockEntity blockEntity) {
+    private static Optional<GameTestInfo> createGameTestInfo(BlockPos testBlockPos, CommandSourceStack source, RetryOptions retryOptions) {
+        ServerLevel level = source.getLevel();
+        if (level.getBlockEntity(testBlockPos) instanceof TestInstanceBlockEntity blockEntity) {
             Optional<Holder.Reference<GameTestInstance>> maybeTest = blockEntity.test()
                 .flatMap(source.registryAccess().lookupOrThrow(Registries.TEST_INSTANCE)::get);
             if (maybeTest.isEmpty()) {
@@ -500,14 +_,12 @@
             } else {
                 Holder.Reference<GameTestInstance> test = maybeTest.get();
                 GameTestInfo testInfo = new GameTestInfo(test, blockEntity.getRotation(), level, retryOptions);
-                testInfo.setTestBlockPos(testBlockPos.pos());
+                testInfo.setTestBlockPos(testBlockPos);
                 return !verifyStructureExists(source, testInfo.getStructure()) ? Optional.empty() : Optional.of(testInfo);
             }
         } else {
             source.sendFailure(
-                Component.translatable(
-                    "commands.test.error.test_instance_not_found.position", testBlockPos.pos().getX(), testBlockPos.pos().getY(), testBlockPos.pos().getZ()
-                )
+                Component.translatable("commands.test.error.test_instance_not_found.position", testBlockPos.getX(), testBlockPos.getY(), testBlockPos.getZ())
             );
             return Optional.empty();
         }
@@ -516,7 +_,7 @@
     private static int createNewStructure(CommandSourceStack source, Identifier id, int xSize, int ySize, int zSize) throws CommandSyntaxException {
         if (xSize <= 48 && ySize <= 48 && zSize <= 48) {
             ServerLevel level = source.getLevel();
-            BlockPos testPos = createTestPositionAround(source, level.dimension());
+            BlockPos testPos = createTestPositionAround(source);
             TestInstanceBlockEntity test = StructureUtils.createNewEmptyTest(id, testPos, new Vec3i(xSize, ySize, zSize), Rotation.NONE, level);
             BlockPos low = test.getStructurePos();
             BlockPos high = low.offset(xSize - 1, 0, zSize - 1);
@@ -590,36 +_,36 @@
         }
     }
 
-    private static BlockPos createTestPositionAround(CommandSourceStack source, ResourceKey<Level> dimension) {
-        TestCommand.Info info = playerAndTestInfo(source, dimension);
+    private static BlockPos createTestPositionAround(CommandSourceStack source) {
+        TestCommand.Info info = playerAndTestInfo(source);
         return new BlockPos(info.playerPos.getX(), info.surfaceY, info.playerPos.getZ() + 3);
     }
 
-    private static TestCommand.Info playerAndTestInfo(CommandSourceStack source, ResourceKey<Level> dimension) {
+    private static TestCommand.Info playerAndTestInfo(CommandSourceStack source) {
         ServerPlayer serverPlayer = source.getPlayer();
         BlockPos playerPos = BlockPos.containing(serverPlayer == null ? source.getPosition() : serverPlayer.position());
         BlockPos testPos = new BlockPos(playerPos.getX(), 384, playerPos.getZ());
-        ServerLevel level = source.getServer().getLevel(dimension);
-        level.getChunk(testPos);
-        int surfaceY = level.getHeightmapPos(Heightmap.Types.WORLD_SURFACE, testPos).getY();
+        source.getLevel().getChunk(testPos);
+        int surfaceY = source.getLevel().getHeightmapPos(Heightmap.Types.WORLD_SURFACE, testPos).getY();
         return new TestCommand.Info(source.getPlayer(), playerPos, testPos, surfaceY);
     }
 
-    private static void outputCoordinates(CommandSourceStack source, Set<ResourceKey<Level>> testDimensions) {
-        TestCommand.Info playerInfo = playerAndTestInfo(source, source.getLevel().dimension());
-        ServerPlayer player = playerInfo.player;
+    private static void outputCoordinates(CommandSourceStack source) {
+        TestCommand.Info info = playerAndTestInfo(source);
+        ServerPlayer player = info.player;
         if (player != null) {
             float playerRotation = player.getYRot();
             float playerPitch = player.getXRot();
-            ResourceKey<Level> playerDimensionKey = player.level().dimension();
-            if (testDimensions.size() != 1 || !testDimensions.contains(playerDimensionKey) || !testDimensions.contains(Level.OVERWORLD)) {
-                String playerDimension = playerDimensionKey.identifier().toString();
+            ResourceKey<Level> dimensionKey = player.level().dimension();
+            String playerDimension = dimensionKey.identifier().toString();
+            String testDimension = source.getLevel().dimension().identifier().toString();
+            if (!testDimension.equals(playerDimension)) {
                 Component playerCoordinates = ComponentUtils.wrapInSquareBrackets(
                         Component.translatable(
                             "test.player.coordinates",
-                            playerInfo.playerPos.getX(),
-                            playerInfo.playerPos.getY(),
-                            playerInfo.playerPos.getZ(),
+                            info.playerPos.getX(),
+                            info.playerPos.getY(),
+                            info.playerPos.getZ(),
                             playerDimension.substring(playerDimension.indexOf(58) + 1)
                         )
                     )
@@ -630,11 +_,11 @@
                                     "/execute in "
                                         + playerDimension
                                         + " run tp @s "
-                                        + playerInfo.playerPos.getX()
-                                        + " "
-                                        + playerInfo.playerPos.getY()
-                                        + " "
-                                        + playerInfo.playerPos.getZ()
+                                        + info.playerPos.getX()
+                                        + " "
+                                        + info.playerPos.getY()
+                                        + " "
+                                        + info.playerPos.getZ()
                                         + " "
                                         + Mth.floor(playerRotation)
                                         + " "
@@ -644,38 +_,33 @@
                             .withHoverEvent(new HoverEvent.ShowText(Component.translatable("chat.coordinates.tooltip")))
                     );
                 source.sendSuccess(() -> playerCoordinates, false);
-
-                for (ResourceKey<Level> testDimensionKey : testDimensions) {
-                    TestCommand.Info testInfo = playerAndTestInfo(source, testDimensionKey);
-                    String testDimension = testDimensionKey.identifier().toString();
-                    Component testCoordinates = ComponentUtils.wrapInSquareBrackets(
-                            Component.translatable(
-                                "test.run.coordinates",
-                                testInfo.playerPos.getX(),
-                                testInfo.surfaceY,
-                                testInfo.playerPos.getZ(),
-                                testDimension.substring(testDimension.indexOf(58) + 1)
-                            )
+                Component testCoordinates = ComponentUtils.wrapInSquareBrackets(
+                        Component.translatable(
+                            "test.run.coordinates",
+                            info.playerPos.getX(),
+                            info.surfaceY,
+                            info.playerPos.getZ(),
+                            testDimension.substring(testDimension.indexOf(58) + 1)
                         )
-                        .withStyle(
-                            s -> s.withColor(ChatFormatting.YELLOW)
-                                .withClickEvent(
-                                    new ClickEvent.SuggestCommand(
-                                        "/execute in "
-                                            + testDimension
-                                            + " run tp @s "
-                                            + testInfo.playerPos.getX()
-                                            + " "
-                                            + testInfo.surfaceY
-                                            + " "
-                                            + testInfo.playerPos.getZ()
-                                            + " 0 0"
-                                    )
+                    )
+                    .withStyle(
+                        s -> s.withColor(ChatFormatting.YELLOW)
+                            .withClickEvent(
+                                new ClickEvent.SuggestCommand(
+                                    "/execute in "
+                                        + testDimension
+                                        + " run tp @s "
+                                        + info.playerPos.getX()
+                                        + " "
+                                        + info.surfaceY
+                                        + " "
+                                        + info.playerPos.getZ()
+                                        + " 0 0"
                                 )
-                                .withHoverEvent(new HoverEvent.ShowText(Component.translatable("chat.coordinates.tooltip")))
-                        );
-                    source.sendSuccess(() -> testCoordinates, false);
-                }
+                            )
+                            .withHoverEvent(new HoverEvent.ShowText(Component.translatable("chat.coordinates.tooltip")))
+                    );
+                source.sendSuccess(() -> testCoordinates, false);
             }
         }
     }
@@ -694,29 +_,18 @@
         }
     }
 
-    public static final class TestSummaryDisplayer implements GameTestListener {
-        private final CommandSourceStack source;
-        private final MultipleTestTracker tracker;
-        private final Set<ResourceKey<Level>> dimensions = new ReferenceArraySet<>();
-
-        public TestSummaryDisplayer(CommandSourceStack source, MultipleTestTracker tracker) {
-            this.source = source;
-            this.tracker = tracker;
-        }
-
+    public record TestSummaryDisplayer(CommandSourceStack source, MultipleTestTracker tracker) implements GameTestListener {
         @Override
         public void testStructureLoaded(GameTestInfo testInfo) {
         }
 
         @Override
         public void testPassed(GameTestInfo testInfo, GameTestRunner runner) {
-            this.dimensions.add(testInfo.getTest().dimension());
             this.showTestSummaryIfAllDone();
         }
 
         @Override
         public void testFailed(GameTestInfo testInfo, GameTestRunner runner) {
-            this.dimensions.add(testInfo.getTest().dimension());
             this.showTestSummaryIfAllDone();
         }
 
@@ -739,7 +_,7 @@
                     this.source.sendSystemMessage(Component.translatable("commands.test.summary.optional_failed", this.tracker.getFailedOptionalCount()));
                 }
 
-                TestCommand.outputCoordinates(this.source, this.dimensions);
+                TestCommand.outputCoordinates(this.source);
             }
         }
     }
