--- a/net/minecraft/world/entity/animal/panda/Panda.java
+++ b/net/minecraft/world/entity/animal/panda/Panda.java
@@ -3,6 +_,7 @@
 import com.mojang.serialization.Codec;
 import java.util.EnumSet;
 import java.util.List;
+import java.util.Objects;
 import java.util.function.IntFunction;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ItemParticleOption;
@@ -70,6 +_,7 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityRemoveEvent;
 import org.jspecify.annotations.Nullable;
 
 public class Panda extends Animal {
@@ -134,6 +_,7 @@
     }
 
     public void sit(boolean value) {
+        if (!new io.papermc.paper.event.entity.EntityToggleSitEvent(this.getBukkitEntity(), value).callEvent()) return; // Paper - Add EntityToggleSitEvent
         this.setFlag(8, value);
     }
 
@@ -518,7 +_,9 @@
 
         for (Panda panda : level.getEntitiesOfClass(Panda.class, this.getBoundingBox().inflate(10.0))) {
             if (!panda.isBaby() && panda.onGround() && !panda.isInWater() && panda.canPerformAction()) {
+                if (new com.destroystokyo.paper.event.entity.EntityJumpEvent(getBukkitLivingEntity()).callEvent()) { // Paper - Entity Jump API
                 panda.jumpFromGround();
+                } else { this.setJumping(false); } // Paper - Entity Jump API; setJumping(false) stops a potential loop
             }
         }
 
@@ -529,13 +_,13 @@
 
     @Override
     protected void pickUpItem(ServerLevel level, ItemEntity entity) {
-        if (this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && canPickUpAndEat(entity)) {
+        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(this, entity, 0, !(this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty() && canPickUpAndEat(entity))).isCancelled()) { // CraftBukkit
             this.onItemPickup(entity);
             ItemStack itemStack = entity.getItem();
             this.setItemSlot(EquipmentSlot.MAINHAND, itemStack);
             this.setGuaranteedDrop(EquipmentSlot.MAINHAND);
             this.take(entity, itemStack.getCount());
-            entity.discard();
+            entity.discard(EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
         }
     }
 
@@ -634,8 +_,9 @@
                 }
 
                 if (!this.level().isClientSide() && this.getAge() == 0 && this.canFallInLove()) {
+                    final ItemStack breedCopy = interactionItemStack.copy(); // Paper - Fix EntityBreedEvent copying
                     this.usePlayerItem(player, hand, interactionItemStack);
-                    this.setInLove(player);
+                    this.setInLove(player, breedCopy); // Paper - Fix EntityBreedEvent copying
                 } else {
                     if (!(this.level() instanceof ServerLevel level) || this.isSitting() || this.isInWater()) {
                         return InteractionResult.PASS;
@@ -867,7 +_,7 @@
         @Override
         protected void alertOther(Mob other, LivingEntity hurtByMob) {
             if (other instanceof Panda && other.isAggressive()) {
-                other.setTarget(hurtByMob);
+                other.setTarget(hurtByMob, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY); // CraftBukkit
             }
         }
     }
