--- a/net/minecraft/client/gui/font/providers/BitmapProvider.java
+++ b/net/minecraft/client/gui/font/providers/BitmapProvider.java
@@ -136,7 +_,7 @@
             Identifier texture = this.file.withPrefix("textures/");
 
             try (InputStream resource = resourceManager.open(texture)) {
-                NativeImage image = NativeImage.read(resource);
+                NativeImage image = NativeImage.read(NativeImage.Format.RGBA, resource);
                 int w = image.getWidth();
                 int h = image.getHeight();
                 int glyphWidth = w / this.codepointGrid[0].length;
