--- a/net/minecraft/client/gui/font/providers/GlyphProviderType.java
+++ b/net/minecraft/client/gui/font/providers/GlyphProviderType.java
@@ -7,10 +_,8 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
-import net.neoforged.fml.common.asm.enumextension.IExtensibleEnum;
-
 @OnlyIn(Dist.CLIENT)
-public enum GlyphProviderType implements StringRepresentable, IExtensibleEnum {
+public enum GlyphProviderType implements StringRepresentable {
     BITMAP("bitmap", BitmapProvider.Definition.CODEC),
     TTF("ttf", TrueTypeGlyphProviderDefinition.CODEC),
     SPACE("space", SpaceProvider.Definition.CODEC),
