--- a/net/minecraft/world/entity/monster/zombie/Husk.java
+++ b/net/minecraft/world/entity/monster/zombie/Husk.java
@@ -61,7 +61,7 @@
         if (flag && this.getMainHandItem().isEmpty() && entity instanceof LivingEntity) {
             float f = serverlevel.getCurrentDifficultyAt(this.blockPosition()).getEffectiveDifficulty();
 
-            ((LivingEntity) entity).addEffect(new MobEffectInstance(MobEffects.HUNGER, 140 * (int) f), this);
+            ((LivingEntity) entity).addEffect(new MobEffectInstance(MobEffects.HUNGER, 140 * (int) f), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
         }
 
         return flag;
