--- a/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.Mirror;
 import net.minecraft.world.level.block.Rotation;
+import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.ChestBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
@@ -105,9 +_,15 @@
         protected void handleDataMarker(String markerId, BlockPos position, ServerLevelAccessor level, RandomSource random, BoundingBox chunkBB) {
             if ("chest".equals(markerId)) {
                 level.setBlockAndUpdate(position, Blocks.AIR.defaultBlockState());
-                if (level.getBlockEntity(position.below()) instanceof ChestBlockEntity chestBlockEntity) {
-                    chestBlockEntity.setLootTable(BuiltInLootTables.IGLOO_CHEST, random.nextLong());
+                // CraftBukkit start - ensure block transformation
+                /*
+                BlockEntity chest = level.getBlockEntity(position.below());
+                if (chest instanceof ChestBlockEntity) {
+                    ((ChestBlockEntity)chest).setLootTable(BuiltInLootTables.IGLOO_CHEST, random.nextLong());
                 }
+                */
+                setCraftLootTable(level, position.below(), random, BuiltInLootTables.IGLOO_CHEST);
+                // CraftBukkit end
             }
         }
 
