--- a/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/IglooPieces.java
@@ -106,10 +_,15 @@
         protected void handleDataMarker(String markerId, BlockPos position, ServerLevelAccessor level, RandomSource random, BoundingBox chunkBB) {
             if ("chest".equals(markerId)) {
                 level.setBlock(position, Blocks.AIR.defaultBlockState(), 3);
+                // CraftBukkit start - ensure block transformation
+                /*
                 BlockEntity chest = level.getBlockEntity(position.below());
                 if (chest instanceof ChestBlockEntity) {
                     ((ChestBlockEntity)chest).setLootTable(BuiltInLootTables.IGLOO_CHEST, random.nextLong());
                 }
+                */
+                setCraftLootTable(level, position.below(), random, BuiltInLootTables.IGLOO_CHEST);
+                // CraftBukkit end
             }
         }
 
