--- a/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
@@ -9,7 +_,7 @@
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.LeverBlock;
-import net.minecraft.world.level.block.RedstoneWireBlock;
+import net.minecraft.world.level.block.RedStoneWireBlock;
 import net.minecraft.world.level.block.RepeaterBlock;
 import net.minecraft.world.level.block.StairBlock;
 import net.minecraft.world.level.block.TripWireBlock;
@@ -211,8 +_,8 @@
             );
             BlockState redstoneWireNS = Blocks.REDSTONE_WIRE
                 .defaultBlockState()
-                .setValue(RedstoneWireBlock.NORTH, RedstoneSide.SIDE)
-                .setValue(RedstoneWireBlock.SOUTH, RedstoneSide.SIDE);
+                .setValue(RedStoneWireBlock.NORTH, RedstoneSide.SIDE)
+                .setValue(RedStoneWireBlock.SOUTH, RedstoneSide.SIDE);
             this.placeBlock(level, redstoneWireNS, 5, -3, 7, chunkBB);
             this.placeBlock(level, redstoneWireNS, 5, -3, 6, chunkBB);
             this.placeBlock(level, redstoneWireNS, 5, -3, 5, chunkBB);
@@ -223,8 +_,8 @@
                 level,
                 Blocks.REDSTONE_WIRE
                     .defaultBlockState()
-                    .setValue(RedstoneWireBlock.NORTH, RedstoneSide.SIDE)
-                    .setValue(RedstoneWireBlock.WEST, RedstoneSide.SIDE),
+                    .setValue(RedStoneWireBlock.NORTH, RedstoneSide.SIDE)
+                    .setValue(RedStoneWireBlock.WEST, RedstoneSide.SIDE),
                 5,
                 -3,
                 1,
@@ -234,8 +_,8 @@
                 level,
                 Blocks.REDSTONE_WIRE
                     .defaultBlockState()
-                    .setValue(RedstoneWireBlock.EAST, RedstoneSide.SIDE)
-                    .setValue(RedstoneWireBlock.WEST, RedstoneSide.SIDE),
+                    .setValue(RedStoneWireBlock.EAST, RedstoneSide.SIDE)
+                    .setValue(RedStoneWireBlock.WEST, RedstoneSide.SIDE),
                 4,
                 -3,
                 1,
@@ -303,8 +_,8 @@
                 level,
                 Blocks.REDSTONE_WIRE
                     .defaultBlockState()
-                    .setValue(RedstoneWireBlock.EAST, RedstoneSide.SIDE)
-                    .setValue(RedstoneWireBlock.WEST, RedstoneSide.SIDE),
+                    .setValue(RedStoneWireBlock.EAST, RedstoneSide.SIDE)
+                    .setValue(RedStoneWireBlock.WEST, RedstoneSide.SIDE),
                 8,
                 -3,
                 6,
@@ -314,8 +_,8 @@
                 level,
                 Blocks.REDSTONE_WIRE
                     .defaultBlockState()
-                    .setValue(RedstoneWireBlock.WEST, RedstoneSide.SIDE)
-                    .setValue(RedstoneWireBlock.SOUTH, RedstoneSide.SIDE),
+                    .setValue(RedStoneWireBlock.WEST, RedstoneSide.SIDE)
+                    .setValue(RedStoneWireBlock.SOUTH, RedstoneSide.SIDE),
                 9,
                 -3,
                 6,
@@ -325,8 +_,8 @@
                 level,
                 Blocks.REDSTONE_WIRE
                     .defaultBlockState()
-                    .setValue(RedstoneWireBlock.NORTH, RedstoneSide.SIDE)
-                    .setValue(RedstoneWireBlock.SOUTH, RedstoneSide.UP),
+                    .setValue(RedStoneWireBlock.NORTH, RedstoneSide.SIDE)
+                    .setValue(RedStoneWireBlock.SOUTH, RedstoneSide.UP),
                 9,
                 -3,
                 5,
@@ -371,10 +_,10 @@
                 level,
                 Blocks.REDSTONE_WIRE
                     .defaultBlockState()
-                    .setValue(RedstoneWireBlock.NORTH, RedstoneSide.SIDE)
-                    .setValue(RedstoneWireBlock.SOUTH, RedstoneSide.SIDE)
-                    .setValue(RedstoneWireBlock.EAST, RedstoneSide.SIDE)
-                    .setValue(RedstoneWireBlock.WEST, RedstoneSide.SIDE),
+                    .setValue(RedStoneWireBlock.NORTH, RedstoneSide.SIDE)
+                    .setValue(RedStoneWireBlock.SOUTH, RedstoneSide.SIDE)
+                    .setValue(RedStoneWireBlock.EAST, RedstoneSide.SIDE)
+                    .setValue(RedStoneWireBlock.WEST, RedstoneSide.SIDE),
                 10,
                 -1,
                 9,
