--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -412,10 +_,15 @@
                         BlockPos pos = this.getWorldPos(1, 0, newZ);
                         if (chunkBB.isInside(pos) && this.isInterior(level, 1, 0, newZ, chunkBB)) {
                             this.hasPlacedSpider = true;
+                            // CraftBukkit start
+                            /*
                             level.setBlock(pos, Blocks.SPAWNER.defaultBlockState(), 2);
                             if (level.getBlockEntity(pos) instanceof SpawnerBlockEntity spawner) {
                                 spawner.setEntityId(EntityType.CAVE_SPIDER, random);
                             }
+                            */
+                            placeCraftSpawner(level, pos, org.bukkit.entity.EntityType.CAVE_SPIDER, 2);
+                            // CraftBukkit end
                         }
                     }
                 }
