--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -425,10 +_,15 @@
                         BlockPos blockpos = this.getWorldPos(1, 0, i2);
                         if (p_227747_.isInside(blockpos) && this.isInterior(p_227743_, 1, 0, i2, p_227747_)) {
                             this.hasPlacedSpider = true;
+                            // CraftBukkit start
+                            /*
                             p_227743_.setBlock(blockpos, Blocks.SPAWNER.defaultBlockState(), 2);
                             if (p_227743_.getBlockEntity(blockpos) instanceof SpawnerBlockEntity spawnerblockentity) {
                                 spawnerblockentity.setEntityId(EntityType.CAVE_SPIDER, p_227746_);
                             }
+                            */
+                            placeCraftSpawner(p_227743_, blockpos, org.bukkit.entity.EntityType.CAVE_SPIDER, 2);
+                            // CraftBukkit end
                         }
                     }
                 }
