--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -1090,10 +_,15 @@
                 BlockPos pos = this.getWorldPos(3, 5, 5);
                 if (chunkBB.isInside(pos)) {
                     this.hasPlacedSpawner = true;
+                    // CraftBukkit start
+                    /*
                     level.setBlock(pos, Blocks.SPAWNER.defaultBlockState(), 2);
                     if (level.getBlockEntity(pos) instanceof SpawnerBlockEntity spawner) {
-                        spawner.setEntityId(EntityTypes.BLAZE, random);
+                        spawner.setEntityId(EntityType.BLAZE, random);
                     }
+                    */
+                    placeCraftSpawner(level, pos, org.bukkit.entity.EntityType.BLAZE, 2);
+                    // CraftBukkit end
                 }
             }
 
