--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFossilPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFossilPieces.java
@@ -14,10 +_,10 @@
 import net.minecraft.world.level.block.Rotation;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
+import net.minecraft.world.level.levelgen.structure.StructurePieceAccessor;
 import net.minecraft.world.level.levelgen.structure.TemplateStructurePiece;
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceSerializationContext;
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceType;
-import net.minecraft.world.level.levelgen.structure.pieces.StructurePiecesBuilder;
 import net.minecraft.world.level.levelgen.structure.templatesystem.BlockIgnoreProcessor;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructurePlaceSettings;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
@@ -40,9 +_,13 @@
         Identifier.withDefaultNamespace("nether_fossils/fossil_14")
     };
 
-    public static void addPieces(StructureTemplateManager structureTemplateManager, StructurePiecesBuilder builder, RandomSource random, BlockPos position) {
+    public static void addPieces(
+        StructureTemplateManager structureTemplateManager, StructurePieceAccessor structurePieceAccessor, RandomSource random, BlockPos position
+    ) {
         Rotation nextRotation = Rotation.getRandom(random);
-        builder.addPiece(new NetherFossilPieces.NetherFossilPiece(structureTemplateManager, Util.getRandom(FOSSILS, random), position, nextRotation));
+        structurePieceAccessor.addPiece(
+            new NetherFossilPieces.NetherFossilPiece(structureTemplateManager, Util.getRandom(FOSSILS, random), position, nextRotation)
+        );
     }
 
     public static class NetherFossilPiece extends TemplateStructurePiece {
