--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -1376,10 +_,15 @@
                 BlockPos blockpos = this.getWorldPos(3, 5, 5);
                 if (p_228366_.isInside(blockpos)) {
                     this.hasPlacedSpawner = true;
+                    // CraftBukkit start
+                    /*
                     p_228362_.setBlock(blockpos, Blocks.SPAWNER.defaultBlockState(), 2);
                     if (p_228362_.getBlockEntity(blockpos) instanceof SpawnerBlockEntity spawnerblockentity) {
                         spawnerblockentity.setEntityId(EntityType.BLAZE, p_228365_);
                     }
+                    */
+                    placeCraftSpawner(p_228362_, blockpos, org.bukkit.entity.EntityType.BLAZE, 2);
+                    // CraftBukkit end
                 }
             }
 
