--- a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentStructure.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentStructure.java
@@ -31,7 +_,8 @@
         int offsetX = context.chunkPos().getBlockX(9);
         int offsetZ = context.chunkPos().getBlockZ(9);
 
-        for (Holder<Biome> biome : context.biomeResolver().getBiomesWithin(offsetX, context.chunkGenerator().getSeaLevel(), offsetZ, 29)) {
+        for (Holder<Biome> biome : context.biomeSource()
+            .getBiomesWithin(offsetX, context.chunkGenerator().getSeaLevel(), offsetZ, 29, context.randomState().sampler())) {
             if (!biome.is(BiomeTags.REQUIRED_OCEAN_MONUMENT_SURROUNDING)) {
                 return Optional.empty();
             }
