--- a/net/minecraft/world/level/levelgen/structure/structures/RuinedPortalStructure.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/RuinedPortalStructure.java
@@ -126,8 +_,14 @@
                         setup.canBeCold()
                             && isCold(
                                 origin,
-                                context.biomeResolver()
-                                    .getNoiseBiome(QuartPos.fromBlock(origin.getX()), QuartPos.fromBlock(origin.getY()), QuartPos.fromBlock(origin.getZ())),
+                                context.chunkGenerator()
+                                    .getBiomeSource()
+                                    .getNoiseBiome(
+                                        QuartPos.fromBlock(origin.getX()),
+                                        QuartPos.fromBlock(origin.getY()),
+                                        QuartPos.fromBlock(origin.getZ()),
+                                        randomState.sampler()
+                                    ),
                                 chunkGenerator.getSeaLevel()
                             ),
                         setup.mossiness(),
