--- a/net/minecraft/world/level/levelgen/structure/structures/ShipwreckPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/ShipwreckPieces.java
@@ -118,7 +_,12 @@
         protected void handleDataMarker(String p_229376_, BlockPos p_229377_, ServerLevelAccessor p_229378_, RandomSource p_229379_, BoundingBox p_229380_) {
             ResourceKey<LootTable> resourcekey = ShipwreckPieces.MARKERS_TO_LOOT.get(p_229376_);
             if (resourcekey != null) {
+                // CraftBukkit start - ensure block transformation
+                /*
                 RandomizableContainer.setBlockEntityLootTable(p_229378_, p_229379_, p_229377_.below(), resourcekey);
+                */
+                setCraftLootTable(p_229378_, p_229377_.below(), p_229379_, resourcekey);
+                // CraftBukkit end
             }
         }
 
