--- a/net/minecraft/world/level/levelgen/structure/structures/ShipwreckPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/ShipwreckPieces.java
@@ -18,10 +_,10 @@
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
+import net.minecraft.world.level.levelgen.structure.StructurePieceAccessor;
 import net.minecraft.world.level.levelgen.structure.TemplateStructurePiece;
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceSerializationContext;
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceType;
-import net.minecraft.world.level.levelgen.structure.pieces.StructurePiecesBuilder;
 import net.minecraft.world.level.levelgen.structure.templatesystem.BlockIgnoreProcessor;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructurePlaceSettings;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
@@ -79,13 +_,13 @@
         StructureTemplateManager structureTemplateManager,
         BlockPos position,
         Rotation rotation,
-        StructurePiecesBuilder builder,
+        StructurePieceAccessor structurePieceAccessor,
         RandomSource random,
         boolean isBeached
     ) {
         Identifier identifier = Util.getRandom(isBeached ? STRUCTURE_LOCATION_BEACHED : STRUCTURE_LOCATION_OCEAN, random);
         ShipwreckPieces.ShipwreckPiece piece = new ShipwreckPieces.ShipwreckPiece(structureTemplateManager, identifier, position, rotation, isBeached);
-        builder.addPiece(piece);
+        structurePieceAccessor.addPiece(piece);
         return piece;
     }
 
@@ -139,7 +_,12 @@
         protected void handleDataMarker(String markerId, BlockPos position, ServerLevelAccessor level, RandomSource random, BoundingBox chunkBB) {
             ResourceKey<LootTable> lootTable = ShipwreckPieces.MARKERS_TO_LOOT.get(markerId);
             if (lootTable != null) {
+                // CraftBukkit start - ensure block transformation
+                /*
                 RandomizableContainer.setBlockEntityLootTable(level, random, position.below(), lootTable);
+                */
+                setCraftLootTable(level, position.below(), random, lootTable);
+                // CraftBukkit end
             }
         }
 
