--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -936,10 +_,9 @@
                 BlockPos blockpos = this.getWorldPos(5, 3, 6);
                 if (p_229639_.isInside(blockpos)) {
                     this.hasPlacedSpawner = true;
-                    p_229635_.setBlock(blockpos, Blocks.SPAWNER.defaultBlockState(), 2);
-                    if (p_229635_.getBlockEntity(blockpos) instanceof SpawnerBlockEntity spawnerblockentity) {
-                        spawnerblockentity.setEntityId(EntityType.SILVERFISH, p_229638_);
-                    }
+                    // CraftBukkit start
+                    placeCraftSpawner(p_229635_, blockpos, org.bukkit.entity.EntityType.SILVERFISH, 2);
+                    // CraftBukkit end
                 }
             }
         }
