--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -865,10 +_,15 @@
                 BlockPos pos = this.getWorldPos(5, 3, 6);
                 if (chunkBB.isInside(pos)) {
                     this.hasPlacedSpawner = true;
+                    // CraftBukkit start
+                    /*
                     level.setBlock(pos, Blocks.SPAWNER.defaultBlockState(), 2);
                     if (level.getBlockEntity(pos) instanceof SpawnerBlockEntity spawner) {
-                        spawner.setEntityId(EntityTypes.SILVERFISH, random);
+                        spawner.setEntityId(EntityType.SILVERFISH, random);
                     }
+                    */
+                    placeCraftSpawner(level, pos, org.bukkit.entity.EntityType.SILVERFISH, 2);
+                    // CraftBukkit end
                 }
             }
         }
