--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -29,9 +_,9 @@
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
 import net.minecraft.world.level.levelgen.structure.StructurePiece;
+import net.minecraft.world.level.levelgen.structure.StructurePieceAccessor;
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceSerializationContext;
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceType;
-import net.minecraft.world.level.levelgen.structure.pieces.StructurePiecesBuilder;
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 import org.jspecify.annotations.Nullable;
 
@@ -98,7 +_,7 @@
 
     private static StrongholdPieces.@Nullable StrongholdPiece findAndCreatePieceFactory(
         Class<? extends StrongholdPieces.StrongholdPiece> pieceClass,
-        StructurePiecesBuilder builder,
+        StructurePieceAccessor structurePieceAccessor,
         RandomSource random,
         int footX,
         int footY,
@@ -108,27 +_,27 @@
     ) {
         StrongholdPieces.StrongholdPiece strongholdPiece = null;
         if (pieceClass == StrongholdPieces.Straight.class) {
-            strongholdPiece = StrongholdPieces.Straight.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.Straight.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.PrisonHall.class) {
-            strongholdPiece = StrongholdPieces.PrisonHall.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.PrisonHall.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.LeftTurn.class) {
-            strongholdPiece = StrongholdPieces.LeftTurn.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.LeftTurn.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.RightTurn.class) {
-            strongholdPiece = StrongholdPieces.RightTurn.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.RightTurn.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.RoomCrossing.class) {
-            strongholdPiece = StrongholdPieces.RoomCrossing.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.RoomCrossing.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.StraightStairsDown.class) {
-            strongholdPiece = StrongholdPieces.StraightStairsDown.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.StraightStairsDown.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.StairsDown.class) {
-            strongholdPiece = StrongholdPieces.StairsDown.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.StairsDown.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.FiveCrossing.class) {
-            strongholdPiece = StrongholdPieces.FiveCrossing.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.FiveCrossing.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.ChestCorridor.class) {
-            strongholdPiece = StrongholdPieces.ChestCorridor.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.ChestCorridor.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.Library.class) {
-            strongholdPiece = StrongholdPieces.Library.createPiece(builder, random, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.Library.createPiece(structurePieceAccessor, random, footX, footY, footZ, direction, depth);
         } else if (pieceClass == StrongholdPieces.PortalRoom.class) {
-            strongholdPiece = StrongholdPieces.PortalRoom.createPiece(builder, footX, footY, footZ, direction, depth);
+            strongholdPiece = StrongholdPieces.PortalRoom.createPiece(structurePieceAccessor, footX, footY, footZ, direction, depth);
         }
 
         return strongholdPiece;
@@ -136,7 +_,7 @@
 
     private static StrongholdPieces.@Nullable StrongholdPiece generatePieceFromSmallDoor(
         StrongholdPieces.StartPiece startPiece,
-        StructurePiecesBuilder builder,
+        StructurePieceAccessor structurePieceAccessor,
         RandomSource random,
         int footX,
         int footY,
@@ -149,7 +_,9 @@
         }
 
         if (imposedPiece != null) {
-            StrongholdPieces.StrongholdPiece strongholdPiece = findAndCreatePieceFactory(imposedPiece, builder, random, footX, footY, footZ, direction, depth);
+            StrongholdPieces.StrongholdPiece strongholdPiece = findAndCreatePieceFactory(
+                imposedPiece, structurePieceAccessor, random, footX, footY, footZ, direction, depth
+            );
             imposedPiece = null;
             if (strongholdPiece != null) {
                 return strongholdPiece;
@@ -170,7 +_,7 @@
                     }
 
                     StrongholdPieces.StrongholdPiece strongholdPiece = findAndCreatePieceFactory(
-                        piece.pieceClass, builder, random, footX, footY, footZ, direction, depth
+                        piece.pieceClass, structurePieceAccessor, random, footX, footY, footZ, direction, depth
                     );
                     if (strongholdPiece != null) {
                         piece.placeCount++;
@@ -185,13 +_,13 @@
             }
         }
 
-        BoundingBox box = StrongholdPieces.FillerCorridor.findPieceBox(builder, random, footX, footY, footZ, direction);
+        BoundingBox box = StrongholdPieces.FillerCorridor.findPieceBox(structurePieceAccessor, random, footX, footY, footZ, direction);
         return box != null && box.minY() > 1 ? new StrongholdPieces.FillerCorridor(depth, box, direction) : null;
     }
 
     private static @Nullable StructurePiece generateAndAddPiece(
         StrongholdPieces.StartPiece startPiece,
-        StructurePiecesBuilder builder,
+        StructurePieceAccessor structurePieceAccessor,
         RandomSource random,
         int footX,
         int footY,
@@ -204,9 +_,9 @@
         }
 
         if (Math.abs(footX - startPiece.getBoundingBox().minX()) <= 112 && Math.abs(footZ - startPiece.getBoundingBox().minZ()) <= 112) {
-            StructurePiece newPiece = generatePieceFromSmallDoor(startPiece, builder, random, footX, footY, footZ, direction, depth + 1);
+            StructurePiece newPiece = generatePieceFromSmallDoor(startPiece, structurePieceAccessor, random, footX, footY, footZ, direction, depth + 1);
             if (newPiece != null) {
-                builder.addPiece(newPiece);
+                structurePieceAccessor.addPiece(newPiece);
                 startPiece.pendingChildren.add(newPiece);
             }
 
@@ -240,15 +_,17 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
-            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 1);
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
+            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 1);
         }
 
         public static StrongholdPieces.@Nullable ChestCorridor createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -1, -1, 0, 5, 5, 7, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.ChestCorridor(genDepth, random, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null
+                ? new StrongholdPieces.ChestCorridor(genDepth, random, box, direction)
+                : null;
         }
 
         @Override
@@ -302,11 +_,11 @@
         }
 
         public static @Nullable BoundingBox findPieceBox(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction
         ) {
             int maxLength = 3;
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -1, -1, 0, 5, 5, 4, direction);
-            StructurePiece collisionPiece = builder.findCollisionPiece(box);
+            StructurePiece collisionPiece = structurePieceAccessor.findCollisionPiece(box);
             if (collisionPiece == null) {
                 return null;
             }
@@ -394,7 +_,7 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
             int zOffA = 3;
             int zOffB = 5;
             Direction orientation = this.getOrientation();
@@ -403,29 +_,31 @@
                 zOffB = 8 - zOffB;
             }
 
-            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, builder, random, 5, 1);
+            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 5, 1);
             if (this.leftLow) {
-                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, builder, random, zOffA, 1);
+                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, zOffA, 1);
             }
 
             if (this.leftHigh) {
-                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, builder, random, zOffB, 7);
+                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, zOffB, 7);
             }
 
             if (this.rightLow) {
-                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, builder, random, zOffA, 1);
+                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, zOffA, 1);
             }
 
             if (this.rightHigh) {
-                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, builder, random, zOffB, 7);
+                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, zOffB, 7);
             }
         }
 
         public static StrongholdPieces.@Nullable FiveCrossing createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -4, -3, 0, 10, 9, 11, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.FiveCrossing(genDepth, random, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null
+                ? new StrongholdPieces.FiveCrossing(genDepth, random, box, direction)
+                : null;
         }
 
         @Override
@@ -510,20 +_,22 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
             Direction orientation = this.getOrientation();
             if (orientation != Direction.NORTH && orientation != Direction.EAST) {
-                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 1);
+                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 1);
             } else {
-                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 1);
+                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 1);
             }
         }
 
         public static StrongholdPieces.@Nullable LeftTurn createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -1, -1, 0, 5, 5, 5, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.LeftTurn(genDepth, random, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null
+                ? new StrongholdPieces.LeftTurn(genDepth, random, box, direction)
+                : null;
         }
 
         @Override
@@ -573,12 +_,12 @@
         }
 
         public static StrongholdPieces.@Nullable Library createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -4, -1, 0, 14, 11, 15, direction);
-            if (!isOkBox(box) || builder.findCollisionPiece(box) != null) {
+            if (!isOkBox(box) || structurePieceAccessor.findCollisionPiece(box) != null) {
                 box = BoundingBox.orientBox(footX, footY, footZ, -4, -1, 0, 14, 6, 15, direction);
-                if (!isOkBox(box) || builder.findCollisionPiece(box) != null) {
+                if (!isOkBox(box) || structurePieceAccessor.findCollisionPiece(box) != null) {
                     return null;
                 }
             }
@@ -757,17 +_,17 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
             if (startPiece != null) {
                 ((StrongholdPieces.StartPiece)startPiece).portalRoomPiece = this;
             }
         }
 
         public static StrongholdPieces.@Nullable PortalRoom createPiece(
-            StructurePiecesBuilder builder, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -4, -1, 0, 11, 8, 16, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.PortalRoom(genDepth, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null ? new StrongholdPieces.PortalRoom(genDepth, box, direction) : null;
         }
 
         @Override
@@ -857,10 +_,15 @@
                 BlockPos pos = this.getWorldPos(5, 3, 6);
                 if (chunkBB.isInside(pos)) {
                     this.hasPlacedSpawner = true;
+                    // CraftBukkit start
+                    /*
                     level.setBlock(pos, Blocks.SPAWNER.defaultBlockState(), 2);
                     if (level.getBlockEntity(pos) instanceof SpawnerBlockEntity spawner) {
-                        spawner.setEntityId(EntityTypes.SILVERFISH, random);
+                        spawner.setEntityId(EntityType.SILVERFISH, random);
                     }
+                    */
+                    placeCraftSpawner(level, pos, org.bukkit.entity.EntityType.SILVERFISH, 2);
+                    // CraftBukkit end
                 }
             }
         }
@@ -882,15 +_,17 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
-            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 1);
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
+            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 1);
         }
 
         public static StrongholdPieces.@Nullable PrisonHall createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -1, -1, 0, 9, 5, 11, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.PrisonHall(genDepth, random, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null
+                ? new StrongholdPieces.PrisonHall(genDepth, random, box, direction)
+                : null;
         }
 
         @Override
@@ -971,20 +_,22 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
             Direction orientation = this.getOrientation();
             if (orientation != Direction.NORTH && orientation != Direction.EAST) {
-                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 1);
+                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 1);
             } else {
-                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 1);
+                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 1);
             }
         }
 
         public static StrongholdPieces.@Nullable RightTurn createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -1, -1, 0, 5, 5, 5, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.RightTurn(genDepth, random, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null
+                ? new StrongholdPieces.RightTurn(genDepth, random, box, direction)
+                : null;
         }
 
         @Override
@@ -1033,17 +_,19 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
-            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, builder, random, 4, 1);
-            this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 4);
-            this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 4);
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
+            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 4, 1);
+            this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 4);
+            this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 4);
         }
 
         public static StrongholdPieces.@Nullable RoomCrossing createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -4, -1, 0, 11, 7, 11, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.RoomCrossing(genDepth, random, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null
+                ? new StrongholdPieces.RoomCrossing(genDepth, random, box, direction)
+                : null;
         }
 
         @Override
@@ -1199,19 +_,21 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
             if (this.isSource) {
                 StrongholdPieces.imposedPiece = StrongholdPieces.FiveCrossing.class;
             }
 
-            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 1);
+            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 1);
         }
 
         public static StrongholdPieces.@Nullable StairsDown createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -1, -7, 0, 5, 11, 5, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.StairsDown(genDepth, random, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null
+                ? new StrongholdPieces.StairsDown(genDepth, random, box, direction)
+                : null;
         }
 
         @Override
@@ -1295,22 +_,24 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
-            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 1);
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
+            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 1);
             if (this.leftChild) {
-                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 2);
+                this.generateSmallDoorChildLeft((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 2);
             }
 
             if (this.rightChild) {
-                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 2);
+                this.generateSmallDoorChildRight((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 2);
             }
         }
 
         public static StrongholdPieces.@Nullable Straight createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -1, -1, 0, 5, 5, 7, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.Straight(genDepth, random, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null
+                ? new StrongholdPieces.Straight(genDepth, random, box, direction)
+                : null;
         }
 
         @Override
@@ -1358,15 +_,17 @@
         }
 
         @Override
-        public void addChildren(StructurePiece startPiece, StructurePiecesBuilder builder, RandomSource random) {
-            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, builder, random, 1, 1);
+        public void addChildren(StructurePiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random) {
+            this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)startPiece, structurePieceAccessor, random, 1, 1);
         }
 
         public static StrongholdPieces.@Nullable StraightStairsDown createPiece(
-            StructurePiecesBuilder builder, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
+            StructurePieceAccessor structurePieceAccessor, RandomSource random, int footX, int footY, int footZ, Direction direction, int genDepth
         ) {
             BoundingBox box = BoundingBox.orientBox(footX, footY, footZ, -1, -7, 0, 5, 11, 8, direction);
-            return isOkBox(box) && builder.findCollisionPiece(box) == null ? new StrongholdPieces.StraightStairsDown(genDepth, random, box, direction) : null;
+            return isOkBox(box) && structurePieceAccessor.findCollisionPiece(box) == null
+                ? new StrongholdPieces.StraightStairsDown(genDepth, random, box, direction)
+                : null;
         }
 
         @Override
@@ -1510,7 +_,7 @@
         }
 
         protected @Nullable StructurePiece generateSmallDoorChildForward(
-            StrongholdPieces.StartPiece startPiece, StructurePiecesBuilder builder, RandomSource random, int xOff, int yOff
+            StrongholdPieces.StartPiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random, int xOff, int yOff
         ) {
             Direction orientation = this.getOrientation();
             if (orientation != null) {
@@ -1518,7 +_,7 @@
                     case NORTH:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.minX() + xOff,
                             this.boundingBox.minY() + yOff,
@@ -1529,7 +_,7 @@
                     case SOUTH:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.minX() + xOff,
                             this.boundingBox.minY() + yOff,
@@ -1540,7 +_,7 @@
                     case WEST:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.minX() - 1,
                             this.boundingBox.minY() + yOff,
@@ -1551,7 +_,7 @@
                     case EAST:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.maxX() + 1,
                             this.boundingBox.minY() + yOff,
@@ -1566,7 +_,7 @@
         }
 
         protected @Nullable StructurePiece generateSmallDoorChildLeft(
-            StrongholdPieces.StartPiece startPiece, StructurePiecesBuilder builder, RandomSource random, int yOff, int zOff
+            StrongholdPieces.StartPiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random, int yOff, int zOff
         ) {
             Direction orientation = this.getOrientation();
             if (orientation != null) {
@@ -1574,7 +_,7 @@
                     case NORTH:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.minX() - 1,
                             this.boundingBox.minY() + yOff,
@@ -1585,7 +_,7 @@
                     case SOUTH:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.minX() - 1,
                             this.boundingBox.minY() + yOff,
@@ -1596,7 +_,7 @@
                     case WEST:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.minX() + zOff,
                             this.boundingBox.minY() + yOff,
@@ -1607,7 +_,7 @@
                     case EAST:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.minX() + zOff,
                             this.boundingBox.minY() + yOff,
@@ -1622,7 +_,7 @@
         }
 
         protected @Nullable StructurePiece generateSmallDoorChildRight(
-            StrongholdPieces.StartPiece startPiece, StructurePiecesBuilder builder, RandomSource random, int yOff, int zOff
+            StrongholdPieces.StartPiece startPiece, StructurePieceAccessor structurePieceAccessor, RandomSource random, int yOff, int zOff
         ) {
             Direction orientation = this.getOrientation();
             if (orientation != null) {
@@ -1630,7 +_,7 @@
                     case NORTH:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.maxX() + 1,
                             this.boundingBox.minY() + yOff,
@@ -1641,7 +_,7 @@
                     case SOUTH:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.maxX() + 1,
                             this.boundingBox.minY() + yOff,
@@ -1652,7 +_,7 @@
                     case WEST:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.minX() + zOff,
                             this.boundingBox.minY() + yOff,
@@ -1663,7 +_,7 @@
                     case EAST:
                         return StrongholdPieces.generateAndAddPiece(
                             startPiece,
-                            builder,
+                            structurePieceAccessor,
                             random,
                             this.boundingBox.minX() + zOff,
                             this.boundingBox.minY() + yOff,
