--- a/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+++ b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
@@ -169,7 +_,7 @@
         HoglinAi.updateActivity(this);
         if (this.isConverting()) {
             this.timeInOverworld++;
-            if (this.timeInOverworld > 300) {
+            if (this.timeInOverworld > 300 && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.ZOGLIN, (timer) -> this.timeInOverworld = timer)) {
                 this.makeSound(SoundEvents.HOGLIN_CONVERTED_TO_ZOMBIFIED);
                 this.finishConversion();
             }
@@ -267,7 +_,10 @@
 
     private void finishConversion() {
         this.convertTo(
-            EntityType.ZOGLIN, ConversionParams.single(this, true, false), p_393201_ -> p_393201_.addEffect(new MobEffectInstance(MobEffects.NAUSEA, 200, 0))
+            EntityType.ZOGLIN, ConversionParams.single(this, true, false), p_393201_ -> {
+                p_393201_.addEffect(new MobEffectInstance(MobEffects.NAUSEA, 200, 0));
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_393201_);
+            }
         );
     }
 
@@ -344,27 +_,27 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return this.level().isClientSide() ? null : HoglinAi.getSoundForCurrentActivity(this).orElse(null);
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_34548_) {
+    public SoundEvent getHurtSound(DamageSource p_34548_) {
         return SoundEvents.HOGLIN_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.HOGLIN_DEATH;
     }
 
     @Override
-    protected SoundEvent getSwimSound() {
+    public SoundEvent getSwimSound() {
         return SoundEvents.HOSTILE_SWIM;
     }
 
     @Override
-    protected SoundEvent getSwimSplashSound() {
+    public SoundEvent getSwimSplashSound() {
         return SoundEvents.HOSTILE_SPLASH;
     }
 
