--- a/net/minecraft/world/entity/monster/hoglin/HoglinBase.java
+++ b/net/minecraft/world/entity/monster/hoglin/HoglinBase.java
@@ -47,6 +_,7 @@
             double d5 = d2 * (double)(p_34646_.level().random.nextFloat() * 0.5F + 0.2F);
             Vec3 vec3 = new Vec3(d3, 0.0, d4).normalize().scale(d5).yRot(f);
             double d6 = d2 * (double)p_34646_.level().random.nextFloat() * 0.5;
+            p_34647_.pushingEntity = p_34646_; // Paper - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
             p_34647_.push(vec3.x, d6, vec3.z);
             p_34647_.hurtMarked = true;
         }
