--- a/net/minecraft/nbt/CompoundTag.java
+++ b/net/minecraft/nbt/CompoundTag.java
@@ -48,13 +_,19 @@
             }
         }
 
+        private static byte readNamedTagType(DataInput input, NbtAccounter accounter) throws IOException {
+            accounter.accountBytes(2);
+            return input.readByte();
+        }
+
         private static CompoundTag loadCompound(DataInput input, NbtAccounter accounter) throws IOException {
             accounter.accountBytes(48L);
-            Map<String, Tag> values = Maps.newHashMap();
+            it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<String, Tag> values = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(8, 0.8f); // Paper - Reduce memory footprint of CompoundTag
 
             byte tagType;
-            while ((tagType = input.readByte()) != 0) {
-                String key = readString(input, accounter);
+            while ((tagType = readNamedTagType(input, accounter)) != 0) {
+                String key = accounter.readUTF(input.readUTF());
+                accounter.accountBytes(4); //Forge: 4 extra bytes for the object allocation.
                 Tag tag = CompoundTag.readNamedTagData(TagTypes.getType(tagType), key, input, accounter);
                 if (values.put(key, tag) == null) {
                     accounter.accountBytes(36L);
@@ -164,7 +_,18 @@
     }
 
     public CompoundTag() {
-        this(new HashMap<>());
+        this(new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(8, 0.8f)); // Paper - Reduce memory footprint of CompoundTag
+    }
+
+    /**
+     * Neo: create a compound tag that is generally suitable to hold the given amount of entries
+     * without needing to resize the internal map.
+     *
+     * @param expectedEntries the expected number of entries that the compound tag will have
+     * @see HashMap#newHashMap(int)
+     */
+    public CompoundTag(int expectedEntries) {
+        this(HashMap.newHashMap(expectedEntries));
     }
 
     @Override
@@ -221,6 +_,7 @@
     }
 
     public @Nullable Tag put(String name, Tag tag) {
+        if (tag == null) throw new IllegalArgumentException("Invalid null NBT value with key " + name);
         return this.tags.put(name, tag);
     }
 
@@ -392,9 +_,16 @@
     }
 
     public CompoundTag copy() {
-        HashMap<String, Tag> newTags = new HashMap<>();
-        this.tags.forEach((key, tag) -> newTags.put(key, tag.copy()));
-        return new CompoundTag(newTags);
+        // Paper start - Reduce memory footprint of CompoundTag
+        it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<String, Tag> ret = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(this.tags.size(), 0.8f);
+        java.util.Iterator<java.util.Map.Entry<String, Tag>> iterator = (this.tags instanceof it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap) ? ((it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap)this.tags).object2ObjectEntrySet().fastIterator() : this.tags.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<String, Tag> entry = iterator.next();
+            ret.put(entry.getKey(), entry.getValue().copy());
+        }
+
+        return new CompoundTag(ret);
+        // Paper end - Reduce memory footprint of CompoundTag
     }
 
     @Override
@@ -533,4 +_,29 @@
     public <T> Optional<T> read(MapCodec<T> codec, DynamicOps<Tag> ops) {
         return codec.decode(ops, ops.getMap(this).getOrThrow()).resultOrPartial(error -> LOGGER.error("Failed to read value ({}): {}", this, error));
     }
+
+    // Paper start - option to read via codec without logging errors
+    // The below methods are 1 to 1 copies of the above-defined read methods without the logging part.
+    // Copying was chosen over overloading the read methods as a boolean parameter to mark a method as quiet
+    // is not intuitive and would require even more overloads.
+    // Not a lot of diff in these methods is expected
+    public <T> Optional<T> readQuiet(String name, Codec<T> codec) {
+        return this.readQuiet(name, codec, NbtOps.INSTANCE);
+    }
+
+    public <T> Optional<T> readQuiet(String name, Codec<T> codec, DynamicOps<Tag> ops) {
+        Tag tag = this.get(name);
+        return tag == null
+                ? Optional.empty()
+                : codec.parse(ops, tag).resultOrPartial();
+    }
+
+    public <T> Optional<T> readQuiet(MapCodec<T> codec) {
+        return this.readQuiet(codec, NbtOps.INSTANCE);
+    }
+
+    public <T> Optional<T> readQuiet(MapCodec<T> codec, DynamicOps<Tag> ops) {
+        return codec.decode(ops, ops.getMap(this).getOrThrow()).resultOrPartial();
+    }
+    // Paper end - option to read via codec without logging errors
 }
