--- a/net/minecraft/nbt/NbtIo.java
+++ b/net/minecraft/nbt/NbtIo.java
@@ -118,6 +_,12 @@
     }
 
     public static CompoundTag read(DataInput input, NbtAccounter accounter) throws IOException {
+        // Spigot start
+        if ( input instanceof io.netty.buffer.ByteBufInputStream )
+        {
+            input = new DataInputStream(new org.spigotmc.LimitStream((InputStream) input, accounter));
+        }
+        // Spigot end
         Tag tag = readUnnamedTag(input, accounter);
         if (tag instanceof CompoundTag) {
             return (CompoundTag)tag;
@@ -179,10 +_,12 @@
     @VisibleForTesting
     public static Tag readUnnamedTag(DataInput input, NbtAccounter accounter) throws IOException {
         byte type = input.readByte();
+        accounter.accountBytes(1); // Forge: Count everything!
         if (type == 0) {
             return EndTag.INSTANCE;
         } else {
-            StringTag.skipString(input);
+            accounter.readUTF(input.readUTF()); //Forge: Count this string.
+            accounter.accountBytes(4); //Forge: 4 extra bytes for the object allocation.
             return readTagSafe(input, accounter, type);
         }
     }
