--- a/net/minecraft/world/level/storage/WorldNBTStorage.java
+++ b/net/minecraft/world/level/storage/WorldNBTStorage.java
@@ -21,6 +21,12 @@
 import net.minecraft.world.entity.player.EntityHuman;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import net.minecraft.core.IRegistryCustom;
+import net.minecraft.server.level.EntityPlayer;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+// CraftBukkit end
+
 public class WorldNBTStorage {
 
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -86,6 +92,23 @@
         return Optional.empty();
     }
 
+    // CraftBukkit start
+    public Optional<NBTTagCompound> load(EntityHuman entityhuman) {
+        return load(entityhuman.nameAndId()).map((nbttagcompound) -> {
+            if (entityhuman instanceof EntityPlayer) {
+                CraftPlayer player = (CraftPlayer) entityhuman.getBukkitEntity();
+                // Only update first played if it is older than the one we have
+                long modified = new File(this.playerDir, entityhuman.getStringUUID() + ".dat").lastModified();
+                if (modified < player.getFirstPlayed()) {
+                    player.setFirstPlayed(modified);
+                }
+            }
+
+            return nbttagcompound;
+        });
+    }
+    // CraftBukkit end
+
     public Optional<NBTTagCompound> load(NameAndId nameandid) {
         Optional<NBTTagCompound> optional = this.load(nameandid, ".dat");
 
@@ -102,4 +125,10 @@
             return nbttagcompound;
         });
     }
+
+    // CraftBukkit start
+    public File getPlayerDir() {
+        return playerDir;
+    }
+    // CraftBukkit end
 }
