--- a/net/minecraft/server/commands/data/DataCommands.java
+++ b/net/minecraft/server/commands/data/DataCommands.java
@@ -17,30 +_,21 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.function.BiConsumer;
-import net.minecraft.commands.CommandBuildContext;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.CompoundTagArgument;
 import net.minecraft.commands.arguments.NbtPathArgument;
 import net.minecraft.commands.arguments.NbtTagArgument;
-import net.minecraft.commands.arguments.ResourceOrIdArgument;
-import net.minecraft.core.Holder;
 import net.minecraft.nbt.CollectionTag;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.EndTag;
-import net.minecraft.nbt.FloatTag;
-import net.minecraft.nbt.IntTag;
 import net.minecraft.nbt.NumericTag;
 import net.minecraft.nbt.PrimitiveTag;
 import net.minecraft.nbt.StringTag;
 import net.minecraft.nbt.Tag;
 import net.minecraft.network.chat.Component;
 import net.minecraft.server.commands.ArgProvider;
-import net.minecraft.server.commands.LootContextSources;
 import net.minecraft.util.Mth;
-import net.minecraft.world.level.storage.loot.LootContext;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
 
 public class DataCommands {
     private static final SimpleCommandExceptionType ERROR_MERGE_UNCHANGED = new SimpleCommandExceptionType(Component.translatable("commands.data.merge.failed"));
@@ -66,7 +_,7 @@
     public static final List<ArgProvider<DataAccessor>> TARGET_PROVIDERS = ArgProvider.buildList("target", ALL_PROVIDERS);
     public static final List<ArgProvider<DataAccessor>> SOURCE_PROVIDERS = ArgProvider.buildList("source", ALL_PROVIDERS);
 
-    public static void register(CommandDispatcher<CommandSourceStack> dispatcher, CommandBuildContext buildContext) {
+    public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
         LiteralArgumentBuilder<CommandSourceStack> root = Commands.literal("data").requires(Commands.hasPermission(Commands.LEVEL_GAMEMASTERS));
 
         for (ArgProvider<DataAccessor> targetProvider : TARGET_PROVIDERS) {
@@ -111,7 +_,6 @@
                 )
                 .then(
                     decorateModification(
-                        buildContext,
                         (parent, rest) -> parent.then(
                                 Commands.literal("insert")
                                     .then(
@@ -188,7 +_,7 @@
     }
 
     private static ArgumentBuilder<CommandSourceStack, ?> decorateModification(
-        CommandBuildContext buildContext, BiConsumer<ArgumentBuilder<CommandSourceStack, ?>, DataCommands.DataManipulatorDecorator> nodeSupplier
+        BiConsumer<ArgumentBuilder<CommandSourceStack, ?>, DataCommands.DataManipulatorDecorator> nodeSupplier
     ) {
         LiteralArgumentBuilder<CommandSourceStack> modify = Commands.literal("modify");
 
@@ -261,32 +_,6 @@
                         );
                     }
 
-                    nodeSupplier.accept(
-                        targetPathNode,
-                        manipulator -> LootContextSources.addContextSources(
-                            Commands.literal("compute"),
-                            (contextDecorator, output) -> {
-                                output.accept(
-                                    Commands.literal("float")
-                                        .then(Commands.argument("provider", ResourceOrIdArgument.floatProvider(buildContext)).executes(c -> {
-                                            LootContext lootContext = contextDecorator.createContext(c);
-                                            Holder<ContextFloatProvider> provider = ResourceOrIdArgument.getFloatProvider(c, "provider");
-                                            float value = provider.value().getFloat(lootContext);
-                                            return manipulateData(c, targetProvider, manipulator, List.of(FloatTag.valueOf(value)));
-                                        }))
-                                );
-                                output.accept(
-                                    Commands.literal("integer")
-                                        .then(Commands.argument("provider", ResourceOrIdArgument.intProvider(buildContext)).executes(c -> {
-                                            LootContext lootContext = contextDecorator.createContext(c);
-                                            Holder<ContextIntProvider> provider = ResourceOrIdArgument.getIntProvider(c, "provider");
-                                            int value = provider.value().getInt(lootContext);
-                                            return manipulateData(c, targetProvider, manipulator, List.of(IntTag.valueOf(value)));
-                                        }))
-                                );
-                            }
-                        )
-                    );
                     nodeSupplier.accept(
                         targetPathNode, manipulator -> Commands.literal("value").then(Commands.argument("value", NbtTagArgument.nbtTag()).executes(c -> {
                             List<Tag> source = Collections.singletonList(NbtTagArgument.getNbtTag(c, "value"));
