--- a/net/minecraft/server/commands/data/EntityDataAccessor.java
+++ b/net/minecraft/server/commands/data/EntityDataAccessor.java
@@ -43,7 +_,6 @@
         try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(this.entity.problemPath(), LOGGER)) {
             this.entity.load(TagValueInput.create(reporter, this.entity.registryAccess(), tag));
             this.entity.setUUID(uuid);
-            this.entity.postDataManipulated();
         }
     }
 
