--- a/net/minecraft/server/commands/data/StorageDataAccessor.java
+++ b/net/minecraft/server/commands/data/StorageDataAccessor.java
@@ -3,7 +_,6 @@
 import com.mojang.brigadier.builder.ArgumentBuilder;
 import com.mojang.brigadier.context.CommandContext;
 import com.mojang.brigadier.suggestion.SuggestionProvider;
-import com.mojang.brigadier.suggestion.SuggestionsBuilder;
 import java.util.Locale;
 import java.util.function.Function;
 import net.minecraft.commands.CommandSourceStack;
