--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -299,7 +_,7 @@
     @Override
     public InteractionResult mobInteract(Player p_316559_, InteractionHand p_316119_) {
         ItemStack itemstack = p_316559_.getItemInHand(p_316119_);
-        if (itemstack.is(Items.BRUSH) && this.brushOffScute(p_316559_, itemstack)) {
+        if (itemstack.canPerformAction(net.neoforged.neoforge.common.ItemAbilities.BRUSH_BRUSH) && this.brushOffScute(p_316559_, itemstack)) {
             itemstack.hurtAndBreak(16, p_316559_, p_316119_.asEquipmentSlot());
             return InteractionResult.SUCCESS;
         } else {
@@ -331,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return this.isScared() ? null : SoundEvents.ARMADILLO_AMBIENT;
     }
 
@@ -341,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.ARMADILLO_DEATH;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_316220_) {
+    public SoundEvent getHurtSound(DamageSource p_316220_) {
         return this.isScared() ? SoundEvents.ARMADILLO_HURT_REDUCED : SoundEvents.ARMADILLO_HURT;
     }
 
