--- a/net/minecraft/world/entity/npc/villager/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/villager/AbstractVillager.java
@@ -49,15 +_,26 @@
 import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.inventory.CraftMerchant;
+import org.bukkit.craftbukkit.inventory.CraftMerchantRecipe;
+import org.bukkit.event.entity.VillagerAcquireTradeEvent;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
 public abstract class AbstractVillager extends AgeableMob implements Npc, Merchant, InventoryCarrier {
+
+    // CraftBukkit start
+    @Override
+    public CraftMerchant getCraftMerchant() {
+        return (org.bukkit.craftbukkit.entity.CraftAbstractVillager) getBukkitEntity();
+    }
+    // CraftBukkit end
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final EntityDataAccessor<Integer> DATA_UNHAPPY_COUNTER = SynchedEntityData.defineId(AbstractVillager.class, EntityDataSerializers.INT);
     private @Nullable Player tradingPlayer;
     protected @Nullable MerchantOffers offers;
-    private final SimpleContainer inventory = new SimpleContainer(8);
+    private final SimpleContainer inventory = new SimpleContainer(8, (org.bukkit.craftbukkit.entity.CraftAbstractVillager) this.getBukkitEntity()); // CraftBukkit add argument
 
     public AbstractVillager(EntityType<? extends AbstractVillager> type, Level level) {
         super(type, level);
@@ -139,6 +_,7 @@
         if (this.tradingPlayer instanceof ServerPlayer) {
             CriteriaTriggers.TRADE.trigger((ServerPlayer)this.tradingPlayer, this, offer.getResult());
         }
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.TradeWithVillagerEvent(this.tradingPlayer, offer, this));
     }
 
     protected abstract void rewardTradeXp(final MerchantOffer offer);
@@ -245,15 +_,15 @@
                 .create(tradeSet.randomSequence());
             int numberOfOffers = tradeSet.calculateNumberOfTrades(lootContext);
             if (tradeSet.allowDuplicates()) {
-                addOffersFromItemListings(lootContext, offers, tradeSet.getTrades(), numberOfOffers);
+                addOffersFromItemListings(lootContext, offers, tradeSet.getTrades(), numberOfOffers, this);
             } else {
-                addOffersFromItemListingsWithoutDuplicates(lootContext, offers, tradeSet.getTrades(), numberOfOffers);
+                addOffersFromItemListingsWithoutDuplicates(lootContext, offers, tradeSet.getTrades(), numberOfOffers, this);
             }
         }
     }
 
     @VisibleForTesting
-    static void addOffersFromItemListings(LootContext lootContext, MerchantOffers merchantOffers, HolderSet<VillagerTrade> potentialOffers, int numberOfOffers) {
+    static void addOffersFromItemListings(LootContext lootContext, MerchantOffers merchantOffers, HolderSet<VillagerTrade> potentialOffers, int numberOfOffers, final AbstractVillager villager) {
         List<Holder<VillagerTrade>> potentialOffersList = Lists.newArrayList(potentialOffers);
         int offersFound = 0;
 
@@ -271,7 +_,7 @@
     }
 
     private static void addOffersFromItemListingsWithoutDuplicates(
-        LootContext lootContext, MerchantOffers merchantOffers, HolderSet<VillagerTrade> potentialOffers, int numberOfOffers
+        LootContext lootContext, MerchantOffers merchantOffers, HolderSet<VillagerTrade> potentialOffers, int numberOfOffers, final AbstractVillager villager
     ) {
         List<Holder<VillagerTrade>> leftoverOffers = Lists.newArrayList(potentialOffers);
         int offersFound = 0;
@@ -280,7 +_,16 @@
             Holder<VillagerTrade> villagerTrade = leftoverOffers.remove(lootContext.getRandom().nextInt(leftoverOffers.size()));
             MerchantOffer offer = villagerTrade.value().getOffer(lootContext);
             if (offer != null) {
-                merchantOffers.add(offer);
+                // CraftBukkit start
+                VillagerAcquireTradeEvent event = new VillagerAcquireTradeEvent((org.bukkit.entity.AbstractVillager) villager.getBukkitEntity(), offer.asBukkit());
+                // Suppress during worldgen
+                if (villager.valid) {
+                    Bukkit.getPluginManager().callEvent(event);
+                }
+                if (!event.isCancelled()) {
+                    merchantOffers.add(CraftMerchantRecipe.fromBukkit(event.getRecipe()).toMinecraft());
+                }
+                // CraftBukkit end
                 offersFound++;
             }
         }
