--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -90,8 +_,11 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.timeline.Timelines;
+import org.bukkit.Bukkit;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.VillagerReplenishTradeEvent;
 
 public class Villager extends AbstractVillager implements VillagerDataHolder, ReputationEventHandler {
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -118,7 +_,7 @@
     private boolean increaseProfessionLevelOnUpdate;
     private @Nullable Player lastTradedPlayer;
     private int foodLevel = 0;
-    private final GossipContainer gossips = new GossipContainer();
+    private final GossipContainer gossips = new GossipContainer(this); // CraftBukkit - add constructor parameter in Reputation
     private long lastGossipTime;
     private long lastGossipDecayTime = 0L;
     private int villagerXp = 0;
@@ -180,6 +_,9 @@
         MemoryModuleType.MEETING_POINT,
         (villager, poiType) -> poiType.is(PoiTypes.MEETING)
     );
+    // CraftBukkit start
+    public long gossipDecayInterval = GOSSIP_DECAY_INTERVAL;
+    // CraftBukkit end
 
     public Villager(EntityType<? extends Villager> entityType, Level level) {
         super(entityType, level);
@@ -252,7 +_,7 @@
                     this.increaseProfessionLevelOnUpdate = false;
                 }
 
-                this.addEffect(new MobEffectInstance(MobEffects.REGENERATION, 200, 0));
+                this.addEffect(new MobEffectInstance(MobEffects.REGENERATION, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.VILLAGER_TRADE); // CraftBukkit
             }
         }
 
@@ -289,7 +_,7 @@
     @Override
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping()) {
+        if (itemStack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping() || player.isSecondaryUseActive()) {
             return super.mobInteract(player, hand);
         }
 
@@ -368,7 +_,13 @@
         this.updateDemand();
 
         for (MerchantOffer offer : this.getOffers()) {
-            offer.resetUses();
+            // CraftBukkit start
+            VillagerReplenishTradeEvent event = new VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), offer.asBukkit());
+            Bukkit.getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+                offer.resetUses();
+            }
+            // CraftBukkit end
         }
 
         this.resendOffersToTradingPlayer();
@@ -429,7 +_,13 @@
         int missedUpdates = 2 - this.numberOfRestocksToday;
         if (missedUpdates > 0) {
             for (MerchantOffer offer : this.getOffers()) {
-                offer.resetUses();
+                // CraftBukkit start
+                VillagerReplenishTradeEvent event = new VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), offer.asBukkit());
+                Bukkit.getPluginManager().callEvent(event);
+                if (!event.isCancelled()) {
+                    offer.resetUses();
+                }
+                // CraftBukkit end
             }
         }
 
@@ -602,7 +_,7 @@
 
     @Override
     public void die(DamageSource source) {
-        LOGGER.info("Villager {} died, message: '{}'", this, source.getLocalizedDeathMessage(this).getString());
+        if (org.spigotmc.SpigotConfig.logVillagerDeaths) LOGGER.info("Villager {} died, message: '{}'", this, source.getLocalizedDeathMessage(this).getString()); // Spigot
         Entity murderer = source.getEntity();
         if (murderer != null) {
             this.tellWitnessesThatIWasMurdered(murderer);
@@ -701,7 +_,7 @@
         return VillagerData.canLevelUp(currentLevel) && this.villagerXp >= VillagerData.getMaxXpPerLevel(currentLevel);
     }
 
-    private void increaseMerchantCareer(ServerLevel level) {
+    public void increaseMerchantCareer(ServerLevel level) {
         this.setVillagerData(this.getVillagerData().withLevel(this.getVillagerData().level() + 1));
         this.updateTrades(level);
     }
@@ -761,13 +_,14 @@
 
     @Override
     public void thunderHit(ServerLevel level, LightningBolt lightningBolt) {
-        if (level.getDifficulty() != Difficulty.PEACEFUL) {
+        if (level.getDifficulty() != Difficulty.PEACEFUL && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityTypes.WITCH, (timer) -> {})) {
             LOGGER.info("Villager {} was struck by lightning {}.", this, lightningBolt);
             Witch witch = this.convertTo(EntityTypes.WITCH, ConversionParams.single(this, false, false), w -> {
                 w.finalizeSpawn(level, level.getCurrentDifficultyAt(w.blockPosition()), EntitySpawnReason.CONVERSION, null);
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, w);
                 w.setPersistenceRequired();
                 this.releaseAllPois();
-            });
+            }, EntityTransformEvent.TransformReason.LIGHTNING, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); // CraftBukkit
             if (witch == null) {
                 super.thunderHit(level, lightningBolt);
             }
@@ -784,7 +_,10 @@
     @Override
     public boolean wantsToPickUp(ServerLevel level, ItemStack itemStack) {
         Item item = itemStack.getItem();
-        return (itemStack.is(ItemTags.VILLAGER_PICKS_UP) || this.getVillagerData().profession().value().requestedItems().contains(item))
+        // Neo: Allow Farmer Villagers to pick up SpecialPlantable items.
+        boolean isFarmerDesiredSeed = (itemStack.getItem() instanceof net.neoforged.neoforge.common.SpecialPlantable specialPlantable && specialPlantable.villagerCanPlantItem(this))
+                && this.getVillagerData().profession().value().secondaryPoi().stream().anyMatch(secondaryPoi -> secondaryPoi.defaultBlockState().is(net.neoforged.neoforge.common.Tags.Blocks.VILLAGER_FARMLANDS));
+        return (itemStack.is(ItemTags.VILLAGER_PICKS_UP) || this.getVillagerData().profession().value().requestedItems().contains(item) || isFarmerDesiredSeed)
             && this.getInventory().canAddItem(itemStack);
     }
 
@@ -806,16 +_,24 @@
     }
 
     @Override
-    protected void updateTrades(ServerLevel level) {
+    public void updateTrades(ServerLevel level) {
+        // Paper start - More vanilla friendly methods to update trades
+        this.updateTrades(level, -1);
+    }
+
+    public boolean updateTrades(final ServerLevel level, int amount) {
+        // Paper end - More vanilla friendly methods to update trades
         VillagerData data = this.getVillagerData();
         VillagerProfession profession = data.profession().value();
         ResourceKey<TradeSet> trades = profession.getTrades(data.level());
         if (trades != null) {
-            this.addOffersFromTradeSet(level, this.getOffers(), trades);
+            this.addOffersFromTradeSet(level, this.getOffers(), trades, amount); // Paper - More vanilla friendly methods to update trades
             if (SharedConstants.DEBUG_UNLOCK_ALL_TRADES && data.level() < 5) {
                 this.increaseMerchantCareer(level);
             }
+            return true; // Paper - More vanilla friendly methods to update trades
         }
+        return false; // Paper - More vanilla friendly methods to update trades
     }
 
     public void gossip(ServerLevel level, Villager target, long timestamp) {
@@ -832,7 +_,7 @@
         long timestamp = this.level().getGameTime();
         if (this.lastGossipDecayTime == 0L) {
             this.lastGossipDecayTime = timestamp;
-        } else if (timestamp >= this.lastGossipDecayTime + 24000L) {
+        } else if (timestamp >= this.lastGossipDecayTime + gossipDecayInterval) { // CraftBukkit - use variable for decay interval
             this.gossips.decay();
             this.lastGossipDecayTime = timestamp;
         }
@@ -856,9 +_,10 @@
                         8,
                         6,
                         SpawnUtil.Strategy.LEGACY_IRON_GOLEM,
-                        false
+                        false,
+                        org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_DEFENSE
                     )
-                    .isEmpty()) {
+                    .isEmpty()) { // CraftBukkit
                     nearbyVillagers.forEach(GolemSensor::golemDetected);
                 }
             }
