--- a/net/minecraft/world/entity/npc/villager/VillagerProfession.java
+++ b/net/minecraft/world/entity/npc/villager/VillagerProfession.java
@@ -179,7 +_,7 @@
             FARMER,
             PoiTypes.FARMER,
             ImmutableSet.of(Items.WHEAT, Items.WHEAT_SEEDS, Items.BEETROOT_SEEDS, Items.BONE_MEAL),
-            ImmutableSet.of(Blocks.FARMLAND),
+            ImmutableSet.of(Blocks.SOUL_SAND), // Purpur - Option for Villager Clerics to farm Nether Wart
             SoundEvents.VILLAGER_WORK_FARMER,
             Int2ObjectMap.ofEntries(
                 Int2ObjectMap.entry(1, TradeSets.FARMER_LEVEL_1),
