--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -46,7 +_,6 @@
 import net.minecraft.world.entity.ConversionParams;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityDimensions;
-import net.minecraft.world.entity.EntityEvent;
 import net.minecraft.world.entity.EntitySpawnReason;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EntityTypes;
@@ -79,7 +_,6 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.raid.Raid;
 import net.minecraft.world.entity.schedule.Activity;
-import net.minecraft.world.food.VillagerFood;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
@@ -92,14 +_,18 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.timeline.Timelines;
+import org.bukkit.Bukkit;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.VillagerReplenishTradeEvent;
 
 public class Villager extends AbstractVillager implements VillagerDataHolder, ReputationEventHandler {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final EntityDataAccessor<VillagerData> DATA_VILLAGER_DATA = SynchedEntityData.defineId(Villager.class, EntityDataSerializers.VILLAGER_DATA);
     private static final EntityDataAccessor<Boolean> DATA_VILLAGER_DATA_FINALIZED = SynchedEntityData.defineId(Villager.class, EntityDataSerializers.BOOLEAN);
     public static final int BREEDING_FOOD_THRESHOLD = 12;
+    public static final Map<Item, Integer> FOOD_POINTS = ImmutableMap.of(Items.BREAD, 4, Items.POTATO, 1, Items.CARROT, 1, Items.BEETROOT, 1);
     private static final int MAX_GOSSIP_TOPICS = 10;
     private static final int GOSSIP_COOLDOWN = 1200;
     private static final int GOSSIP_DECAY_INTERVAL = 24000;
@@ -113,16 +_,20 @@
     private static final int DEFAULT_LAST_RESTOCK = 0;
     private static final int DEFAULT_LAST_GOSSIP_DECAY = 0;
     private static final int DEFAULT_RESTOCKS_TODAY = 0;
+    private static final boolean DEFAULT_ASSIGN_PROFESSION_WHEN_SPAWNED = false;
     private static final EntityDimensions BABY_DIMENSIONS = EntityDimensions.scalable(0.49F, 0.98F).withEyeHeight(0.63F);
+    private int updateMerchantTimer;
+    private boolean increaseProfessionLevelOnUpdate;
     private @Nullable Player lastTradedPlayer;
     private int foodLevel = 0;
-    private final GossipContainer gossips = new GossipContainer();
+    private final GossipContainer gossips = new GossipContainer(this); // CraftBukkit - add constructor parameter in Reputation
     private long lastGossipTime;
     private long lastGossipDecayTime = 0L;
     private int villagerXp = 0;
     private long lastRestockGameTime = 0L;
     public int numberOfRestocksToday = 0;
     private long lastRestockCheckDay;
+    private boolean assignProfessionWhenSpawned = false;
     private static final Brain.Provider<Villager> BRAIN_PROVIDER = Brain.<Villager>provider(
         List.of(
             SensorType.NEAREST_LIVING_ENTITIES,
@@ -144,7 +_,7 @@
                 activities.add(
                     ActivityData.create(
                         Activity.WORK,
-                        VillagerGoalPackages.getWorkPackage(profession, 0.5F),
+                        VillagerGoalPackages.getWorkPackage(profession, 0.5F, body.level().purpurConfig.villagerClericsFarmWarts), // Purpur - Option for Villager Clerics to farm Nether Wart
                         ImmutableSet.of(Pair.of(MemoryModuleType.JOB_SITE, MemoryStatus.VALUE_PRESENT))
                     )
                 );
@@ -177,6 +_,8 @@
         MemoryModuleType.MEETING_POINT,
         (villager, poiType) -> poiType.is(PoiTypes.MEETING)
     );
+    private boolean isLobotomized = false; public boolean isLobotomized() { return this.isLobotomized; } // Purpur - Lobotomize stuck villagers
+    private int notLobotomizedCount = 0; // Purpur - Lobotomize stuck villagers
 
     public Villager(EntityType<? extends Villager> entityType, Level level) {
         super(entityType, level);
@@ -186,6 +_,110 @@
         this.setCanPickUpLoot(true);
     }
 
+    // Purpur start - Allow leashing villagers
+    @Override
+    public boolean canBeLeashed() {
+        return level().purpurConfig.villagerCanBeLeashed;
+    }
+    // Purpur end - Allow leashing villagers
+
+    // Purpur start - Lobotomize stuck villagers
+    private boolean checkLobotomized() {
+        int interval = this.level().purpurConfig.villagerLobotomizeCheckInterval;
+        boolean shouldCheckForTradeLocked = this.level().purpurConfig.villagerLobotomizeWaitUntilTradeLocked;
+        if (this.notLobotomizedCount > 3) {
+            // check half as often if not lobotomized for the last 3+ consecutive checks
+            interval *= 2;
+        }
+        if (this.level().getGameTime() % interval == 0) {
+            // offset Y for short blocks like dirt_path/farmland
+            this.isLobotomized = !(shouldCheckForTradeLocked && this.getVillagerXp() == 0) && !canTravelFrom(net.minecraft.core.BlockPos.containing(this.position().x, this.getBoundingBox().minY + 0.0625D, this.position().z));
+
+            if (this.isLobotomized) {
+                this.notLobotomizedCount = 0;
+            } else {
+                this.notLobotomizedCount++;
+            }
+        }
+        return this.isLobotomized;
+    }
+
+    private boolean canTravelFrom(net.minecraft.core.BlockPos pos) {
+        return canTravelTo(pos.east()) || canTravelTo(pos.west()) || canTravelTo(pos.north()) || canTravelTo(pos.south());
+    }
+
+    private boolean canTravelTo(net.minecraft.core.BlockPos pos) {
+        net.minecraft.world.level.block.state.BlockState state = this.level().getBlockStateIfLoaded(pos);
+        if (state == null) {
+            // chunk not loaded
+            return false;
+        }
+        net.minecraft.world.level.block.Block bottom = state.getBlock();
+        if (bottom instanceof net.minecraft.world.level.block.FenceBlock ||
+                bottom instanceof net.minecraft.world.level.block.FenceGateBlock ||
+                bottom instanceof net.minecraft.world.level.block.WallBlock) {
+            // bottom block is too tall to get over
+            return false;
+        }
+        net.minecraft.world.level.block.Block top = level().getBlockState(pos.above()).getBlock();
+        // only if both blocks have no collision
+        return !bottom.hasCollision && !top.hasCollision;
+    }
+    // Purpur end - Lobotomize stuck villagers
+
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.villagerRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.villagerRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.villagerControllable;
+    }
+
+    @Override
+    protected void registerGoals() {
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this));
+        if (level().purpurConfig.villagerFollowEmeraldBlock) this.goalSelector.addGoal(3, new net.minecraft.world.entity.ai.goal.TemptGoal(this, 1.0D, TEMPT_ITEMS, false)); // Purpur - Villagers follow emerald blocks
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.safeSetAttribute(Attributes.MAX_HEALTH, this.level().purpurConfig.villagerMaxHealth);
+        this.safeSetAttribute(Attributes.SCALE, this.level().purpurConfig.villagerScale);
+        this.safeSetAttribute(Attributes.TEMPT_RANGE, this.level().purpurConfig.villagerTemptRange); // Purpur - Villagers follow emerald blocks
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.villagerTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.villagerAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
+    // Purpur start - Check mobGriefing Overrides
+    @Override
+    public Boolean checkEntityPickUpLootOverride() {
+        return this.level().purpurConfig.villagerCanPickUpLoot;
+    }
+    // Purpur end - Check mobGriefing Overrides
+
     @Override
     public Brain<Villager> getBrain() {
         return (Brain<Villager>) super.getBrain();
@@ -224,32 +_,88 @@
     }
 
     public static AttributeSupplier.Builder createAttributes() {
-        return Mob.createMobAttributes().add(Attributes.MOVEMENT_SPEED, 0.5);
-    }
+        return Mob.createMobAttributes().add(Attributes.MOVEMENT_SPEED, 0.5).add(Attributes.TEMPT_RANGE, 10.0D); // Purpur - Villagers follow emerald blocks
+    }
+
+    public boolean assignProfessionWhenSpawned() {
+        return this.assignProfessionWhenSpawned;
+    }
+
+    // Paper start - EAR 2
+    @Override
+    public void inactiveTick() {
+        // SPIGOT-3874, SPIGOT-3894, SPIGOT-3846, SPIGOT-5286 :(
+        if (this.getUnhappyCounter() > 0) {
+            this.setUnhappyCounter(this.getUnhappyCounter() - 1);
+        }
+        if (this.aware && this.isEffectiveAi()) { // Paper - Do not tick AI for inactive unaware mobs
+            if (this.level().spigotConfig.tickInactiveVillagers) {
+                this.customServerAiStep(this.level().getMinecraftWorld());
+            } else {
+                this.customServerAiStep(this.level().getMinecraftWorld(), true);
+            }
+        }
+        this.maybeDecayGossip();
+        super.inactiveTick();
+    }
+    // Paper end - EAR 2
 
     @Override
     protected void customServerAiStep(ServerLevel level) {
+        // Paper start - EAR 2
+        this.customServerAiStep(level, false);
+    }
+    protected void customServerAiStep(ServerLevel level, boolean inactive) { // Purpur - Lobotomize stuck villagers
+        // Paper end - EAR 2
         ProfilerFiller profiler = Profiler.get();
         profiler.push("villagerBrain");
-        this.getBrain().tick(level, this);
+        // Purpur start - Lobotomize stuck villagers
+        if (this.level().purpurConfig.villagerLobotomizeEnabled) {
+            // treat as inactive if lobotomized
+            inactive = inactive || checkLobotomized();
+        } else {
+            this.isLobotomized = false;
+        }
+        if (!inactive && (getRider() == null || !this.isControllable())) { // Purpur - Ridables
+            this.getBrain().tick(level, this); // Paper - EAR 2
+        }
+        else if (this.isLobotomized && shouldRestock(level)) restock();
+        // Purpur end - Lobotomize stuck villagers
         profiler.pop();
+        if (this.assignProfessionWhenSpawned) {
+            this.assignProfessionWhenSpawned = false;
+        }
+
+        if (!this.isTrading() && this.updateMerchantTimer > 0) {
+            this.updateMerchantTimer--;
+            if (this.updateMerchantTimer <= 0) {
+                if (this.increaseProfessionLevelOnUpdate) {
+                    this.increaseMerchantCareer(level);
+                    this.increaseProfessionLevelOnUpdate = false;
+                }
+
+                this.addEffect(new MobEffectInstance(MobEffects.REGENERATION, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.VILLAGER_TRADE); // CraftBukkit
+            }
+        }
+
         if (this.lastTradedPlayer != null) {
             level.onReputationEvent(ReputationEventType.TRADE, this.lastTradedPlayer, this);
             level.broadcastEntityEvent(this, (byte)14);
             this.lastTradedPlayer = null;
         }
 
-        if (!this.isNoAi() && this.random.nextInt(100) == 0) {
+        if (!inactive && !this.isNoAi() && this.random.nextInt(100) == 0) { // Paper - EAR 2
             Raid raid = level.getRaidAt(this.blockPosition());
             if (raid != null && raid.isActive() && !raid.isOver()) {
                 level.broadcastEntityEvent(this, (byte)42);
             }
         }
 
-        if ((this.offers == null || this.offers.isEmpty()) && this.isTrading()) {
+        if (this.getVillagerData().profession().is(VillagerProfession.NONE) && this.isTrading()) {
             this.stopTrading();
         }
 
+        if (inactive) return; // Paper - EAR 2
         super.customServerAiStep(level);
     }
 
@@ -266,7 +_,7 @@
     @Override
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping()) {
+        if (itemStack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping() || player.isSecondaryUseActive()) {
             return super.mobInteract(player, hand);
         }
 
@@ -289,6 +_,9 @@
                 return InteractionResult.CONSUME;
             }
 
+            if (level().purpurConfig.villagerRidable && itemStack.isEmpty()) return tryRide(player, hand); // Purpur - Ridables
+
+            if (this.level().purpurConfig.villagerAllowTrading) // Purpur - Add config for villager trading
             this.startTrading(player);
         }
 
@@ -317,6 +_,12 @@
         }
     }
 
+    @Override
+    protected void stopTrading() {
+        super.stopTrading();
+        this.resetSpecialPrices();
+    }
+
     private void resetSpecialPrices() {
         if (!this.level().isClientSide()) {
             for (MerchantOffer offer : this.getOffers()) {
@@ -339,7 +_,13 @@
         this.updateDemand();
 
         for (MerchantOffer offer : this.getOffers()) {
-            offer.resetUses();
+            // CraftBukkit start
+            VillagerReplenishTradeEvent event = new VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), offer.asBukkit());
+            Bukkit.getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+                offer.resetUses();
+            }
+            // CraftBukkit end
         }
 
         this.resendOffersToTradingPlayer();
@@ -400,7 +_,13 @@
         int missedUpdates = 2 - this.numberOfRestocksToday;
         if (missedUpdates > 0) {
             for (MerchantOffer offer : this.getOffers()) {
-                offer.resetUses();
+                // CraftBukkit start
+                VillagerReplenishTradeEvent event = new VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), offer.asBukkit());
+                Bukkit.getPluginManager().callEvent(event);
+                if (!event.isCancelled()) {
+                    offer.resetUses();
+                }
+                // CraftBukkit end
             }
         }
 
@@ -413,23 +_,27 @@
 
     public void updateDemand() {
         for (MerchantOffer offer : this.getOffers()) {
-            offer.updateDemand();
+            offer.updateDemand(this.level().purpurConfig.villagerMinimumDemand); // Purpur - Configurable minimum demand for trades
         }
     }
 
-    public void updateSpecialPrices(Player player) {
-        this.resetSpecialPrices();
+    private void updateSpecialPrices(Player player) {
         int reputation = this.getPlayerReputation(player);
-        MobEffectInstance heroOfTheVillage = player.getEffect(MobEffects.HERO_OF_THE_VILLAGE);
-        double heroModifier = heroOfTheVillage == null ? 0.0 : 0.3F + 0.0625F * heroOfTheVillage.getAmplifier();
-
-        for (MerchantOffer offer : this.getOffers()) {
-            if (reputation != 0) {
+        if (reputation != 0) {
+            for (MerchantOffer offer : this.getOffers()) {
+                if (offer.ignoreDiscounts) continue; // Paper - Add ignore discounts API
                 offer.addToSpecialPriceDiff(-Mth.floor(reputation * offer.getPriceMultiplier()));
             }
-
-            if (heroModifier > 0.0) {
-                int costReduction = (int)Math.floor(heroModifier * offer.getBaseCostA().getCount());
+        }
+
+        if (player.hasEffect(MobEffects.HERO_OF_THE_VILLAGE)) {
+            MobEffectInstance effect = player.getEffect(MobEffects.HERO_OF_THE_VILLAGE);
+            int amplifier = effect.getAmplifier();
+
+            for (MerchantOffer offer : this.getOffers()) {
+                if (offer.ignoreDiscounts) continue; // Paper - Add ignore discounts API
+                double modifier = 0.3 + 0.0625 * amplifier;
+                int costReduction = (int)Math.floor(modifier * offer.getBaseCostA().getCount());
                 offer.addToSpecialPriceDiff(-Math.max(costReduction, 1));
             }
         }
@@ -459,6 +_,9 @@
         output.putLong("LastRestock", this.lastRestockGameTime);
         output.putLong("LastGossipDecay", this.lastGossipDecayTime);
         output.putInt("RestocksToday", this.numberOfRestocksToday);
+        if (this.assignProfessionWhenSpawned) {
+            output.putBoolean("AssignProfessionWhenSpawned", true);
+        }
     }
 
     @Override
@@ -477,17 +_,12 @@
         this.villagerXp = input.getIntOr("Xp", 0);
         this.lastRestockGameTime = input.getLongOr("LastRestock", 0L);
         this.lastGossipDecayTime = input.getLongOr("LastGossipDecay", 0L);
+        if (this.level() instanceof ServerLevel) {
+            this.refreshBrain((ServerLevel)this.level());
+        }
+
         this.numberOfRestocksToday = input.getIntOr("RestocksToday", 0);
-        if (this.level() instanceof ServerLevel serverLevel) {
-            this.refreshBrain(serverLevel);
-        }
-    }
-
-    @Override
-    public void postDataManipulated() {
-        if (this.isTrading()) {
-            this.resendOffersToTradingPlayer();
-        }
+        this.assignProfessionWhenSpawned = input.getBooleanOr("AssignProfessionWhenSpawned", false);
     }
 
     @Override
@@ -549,16 +_,13 @@
         this.villagerXp = this.villagerXp + offer.getXp();
         this.lastTradedPlayer = this.getTradingPlayer();
         if (this.shouldIncreaseLevel()) {
-            if (this.level() instanceof ServerLevel serverLevel) {
-                this.increaseMerchantCareer(serverLevel);
-                this.addEffect(new MobEffectInstance(MobEffects.REGENERATION, 200, 0));
-            }
-
+            this.updateMerchantTimer = 40;
+            this.increaseProfessionLevelOnUpdate = true;
             popXp += 5;
         }
 
         if (offer.shouldRewardExp()) {
-            this.level().addFreshEntity(new ExperienceOrb(this.level(), this.getX(), this.getY() + 0.5, this.getZ(), popXp));
+            this.level().addFreshEntity(new ExperienceOrb(this.level(), this.getX(), this.getY() + 0.5, this.getZ(), popXp, org.bukkit.entity.ExperienceOrb.SpawnReason.VILLAGER_TRADE, this.getTradingPlayer(), this)); // Paper
         }
     }
 
@@ -576,7 +_,7 @@
 
     @Override
     public void die(DamageSource source) {
-        LOGGER.info("Villager {} died, message: '{}'", this, source.getLocalizedDeathMessage(this).getString());
+        if (org.spigotmc.SpigotConfig.logVillagerDeaths) LOGGER.info("Villager {} died, message: '{}'", this, source.getLocalizedDeathMessage(this).getString()); // Spigot
         Entity murderer = source.getEntity();
         if (murderer != null) {
             this.tellWitnessesThatIWasMurdered(murderer);
@@ -624,7 +_,7 @@
 
     @Override
     public boolean canBreed() {
-        return this.foodLevel + this.countFoodPointsInInventory() >= 12 && !this.isSleeping() && this.getAge() == 0;
+        return this.level().purpurConfig.villagerCanBreed && this.foodLevel + this.countFoodPointsInInventory() >= 12 && !this.isSleeping() && this.getAge() == 0; // Purpur - Configurable villager breeding
     }
 
     private boolean hungry() {
@@ -632,24 +_,22 @@
     }
 
     private void eatUntilFull() {
-        if (this.hungry()) {
+        if (this.hungry() && this.countFoodPointsInInventory() != 0) {
             for (int slot = 0; slot < this.getInventory().getContainerSize(); slot++) {
                 ItemStack itemStack = this.getInventory().getItem(slot);
-                VillagerFood villagerFood = itemStack.get(DataComponents.VILLAGER_FOOD);
-                if (villagerFood != null) {
-                    int itemCount = itemStack.getCount();
-                    int toRemove = 0;
+                if (!itemStack.isEmpty()) {
+                    Integer value = FOOD_POINTS.get(itemStack.getItem());
+                    if (value != null) {
+                        int itemCount = itemStack.getCount();
 
-                    for (int count = itemCount; count > 0; count--) {
-                        this.foodLevel = this.foodLevel + villagerFood.nutrition();
-                        toRemove++;
-                        if (!this.hungry()) {
-                            this.getInventory().removeItem(slot, toRemove);
-                            return;
+                        for (int count = itemCount; count > 0; count--) {
+                            this.foodLevel = this.foodLevel + value;
+                            this.getInventory().removeItem(slot, 1);
+                            if (!this.hungry()) {
+                                return;
+                            }
                         }
                     }
-
-                    this.getInventory().removeItem(slot, toRemove);
                 }
             }
         }
@@ -677,7 +_,7 @@
         return VillagerData.canLevelUp(currentLevel) && this.villagerXp >= VillagerData.getMaxXpPerLevel(currentLevel);
     }
 
-    private void increaseMerchantCareer(ServerLevel level) {
+    public void increaseMerchantCareer(ServerLevel level) {
         this.setVillagerData(this.getVillagerData().withLevel(this.getVillagerData().level() + 1));
         this.updateTrades(level);
     }
@@ -688,7 +_,7 @@
     }
 
     @Override
-    public void handleEntityEvent(@EntityEvent.Value byte id) {
+    public void handleEntityEvent(byte id) {
         if (id == 12) {
             this.addParticlesAroundSelf(ParticleTypes.HEART);
         } else if (id == 13) {
@@ -711,6 +_,10 @@
         }
 
         this.finalizeVillagerType(level, this.blockPosition());
+        if (spawnReason == EntitySpawnReason.STRUCTURE) {
+            this.assignProfessionWhenSpawned = true;
+        }
+
         return super.finalizeSpawn(level, difficulty, spawnReason, groupData);
     }
 
@@ -733,13 +_,21 @@
 
     @Override
     public void thunderHit(ServerLevel level, LightningBolt lightningBolt) {
-        if (level.getDifficulty() != Difficulty.PEACEFUL) {
-            LOGGER.info("Villager {} was struck by lightning {}.", this, lightningBolt);
+        if (level.getDifficulty() != Difficulty.PEACEFUL && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityTypes.WITCH, (timer) -> {})) {
+            // Paper - Add EntityZapEvent; move log down, event can cancel
             Witch witch = this.convertTo(EntityTypes.WITCH, ConversionParams.single(this, false, false), w -> {
+                // Paper start - Add EntityZapEvent
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityZapEvent(this, lightningBolt, w).isCancelled()) {
+                    return false;
+                }
+                if (org.spigotmc.SpigotConfig.logVillagerDeaths) LOGGER.info("Villager {} was struck by lightning {}.", this, lightningBolt); // Move down
+                // Paper end - Add EntityZapEvent
                 w.finalizeSpawn(level, level.getCurrentDifficultyAt(w.blockPosition()), EntitySpawnReason.CONVERSION, null);
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, w);
                 w.setPersistenceRequired();
                 this.releaseAllPois();
-            });
+                return true; // Paper - Add EntityZapEvent
+            }, EntityTransformEvent.TransformReason.LIGHTNING, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); // CraftBukkit
             if (witch == null) {
                 super.thunderHit(level, lightningBolt);
             }
@@ -756,11 +_,10 @@
     @Override
     public boolean wantsToPickUp(ServerLevel level, ItemStack itemStack) {
         Item item = itemStack.getItem();
-        return (
-                itemStack.is(ItemTags.VILLAGER_PICKS_UP)
-                    || itemStack.has(DataComponents.VILLAGER_FOOD)
-                    || this.getVillagerData().profession().value().requestedItems().contains(item)
-            )
+        // Neo: Allow Farmer Villagers to pick up SpecialPlantable items.
+        boolean isFarmerDesiredSeed = (itemStack.getItem() instanceof net.neoforged.neoforge.common.SpecialPlantable specialPlantable && specialPlantable.villagerCanPlantItem(this))
+                && this.getVillagerData().profession().value().secondaryPoi().stream().anyMatch(secondaryPoi -> secondaryPoi.defaultBlockState().is(net.neoforged.neoforge.common.Tags.Blocks.VILLAGER_FARMLANDS));
+        return (itemStack.is(ItemTags.VILLAGER_PICKS_UP) || this.getVillagerData().profession().value().requestedItems().contains(item) || isFarmerDesiredSeed)
             && this.getInventory().canAddItem(itemStack);
     }
 
@@ -774,53 +_,41 @@
 
     private int countFoodPointsInInventory() {
         SimpleContainer inventory = this.getInventory();
-        int points = 0;
-
-        for (ItemStack item : inventory.getItems()) {
-            VillagerFood villagerFood = item.get(DataComponents.VILLAGER_FOOD);
-            if (villagerFood != null) {
-                points += item.count() * villagerFood.nutrition();
-            }
-        }
-
-        return points;
+        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
     }
 
     public boolean hasFarmSeeds() {
-        return this.getInventory().hasAnyMatching(item -> item.is(ItemTags.VILLAGER_PLANTABLE_SEEDS));
+        return this.getInventory().hasAnyMatching(item -> this.level().purpurConfig.villagerClericsFarmWarts && this.getVillagerData().profession().is(VillagerProfession.CLERIC) ? item.is(Items.NETHER_WART) : item.is(ItemTags.VILLAGER_PLANTABLE_SEEDS));
     }
 
     @Override
-    protected void updateTrades(ServerLevel level) {
+    public void updateTrades(ServerLevel level) {
+        // Paper start - More vanilla friendly methods to update trades
+        this.updateTrades(level, -1);
+    }
+
+    public boolean updateTrades(final ServerLevel level, int amount) {
+        // Paper end - More vanilla friendly methods to update trades
         VillagerData data = this.getVillagerData();
         VillagerProfession profession = data.profession().value();
         ResourceKey<TradeSet> trades = profession.getTrades(data.level());
         if (trades != null) {
-            this.addOffersFromTradeSet(level, this.getOffers(), trades);
+            this.addOffersFromTradeSet(level, this.getOffers(), trades, amount); // Paper - More vanilla friendly methods to update trades
             if (SharedConstants.DEBUG_UNLOCK_ALL_TRADES && data.level() < 5) {
                 this.increaseMerchantCareer(level);
             }
-
-            Player tradingPlayer = this.getTradingPlayer();
-            if (tradingPlayer != null) {
-                this.updateSpecialPrices(tradingPlayer);
-                this.resendOffersToTradingPlayer();
-            }
+            return true; // Paper - More vanilla friendly methods to update trades
         }
+        return false; // Paper - More vanilla friendly methods to update trades
     }
 
     public void gossip(ServerLevel level, Villager target, long timestamp) {
         if ((timestamp < this.lastGossipTime || timestamp >= this.lastGossipTime + 1200L)
             && (timestamp < target.lastGossipTime || timestamp >= target.lastGossipTime + 1200L)) {
-            int newGossip = this.gossips.transferFrom(target.gossips, this.random, 10);
+            this.gossips.transferFrom(target.gossips, this.random, 10);
             this.lastGossipTime = timestamp;
             target.lastGossipTime = timestamp;
             this.spawnGolemIfNeeded(level, timestamp, 5);
-            Player tradingPlayer = this.getTradingPlayer();
-            if (tradingPlayer != null && newGossip > 0) {
-                this.updateSpecialPrices(tradingPlayer);
-                this.resendOffersToTradingPlayer();
-            }
         }
     }
 
@@ -835,6 +_,7 @@
     }
 
     public void spawnGolemIfNeeded(ServerLevel level, long timestamp, int villagersNeededToAgree) {
+        if (level.purpurConfig.villagerSpawnIronGolemRadius > 0 && level.getEntitiesOfClass(net.minecraft.world.entity.animal.golem.IronGolem.class, getBoundingBox().inflate(level.purpurConfig.villagerSpawnIronGolemRadius)).size() > level.purpurConfig.villagerSpawnIronGolemLimit) return; // Purpur - Implement configurable search radius for villagers to spawn iron golems
         if (this.wantsToSpawnGolem(timestamp)) {
             AABB villagerSearchBox = this.getBoundingBox().inflate(10.0, 10.0, 10.0);
             List<Villager> nearbyVillagers = level.getEntitiesOfClass(Villager.class, villagerSearchBox);
@@ -843,7 +_,7 @@
                 .limit(5L)
                 .toList();
             if (nearbyVillagersThatWantAGolem.size() >= villagersNeededToAgree) {
-                if (!SpawnUtil.trySpawnMob(
+                if (!SpawnUtil.trySpawnMob( // Paper - Set Golem Last Seen to stop it from spawning another one - switch to isPresent
                         EntityTypes.IRON_GOLEM,
                         EntitySpawnReason.MOB_SUMMONED,
                         level,
@@ -852,9 +_,11 @@
                         8,
                         6,
                         SpawnUtil.Strategy.LEGACY_IRON_GOLEM,
-                        false
+                        false,
+                        org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_DEFENSE, // CraftBukkit,
+                        () -> {GolemSensor.golemDetected(this);} // Paper - Set Golem Last Seen to stop it from spawning another one
                     )
-                    .isEmpty()) {
+                    .isPresent()) { // Paper - Set Golem Last Seen to stop it from spawning another one - switch to isPresent
                     nearbyVillagers.forEach(GolemSensor::golemDetected);
                 }
             }
@@ -876,12 +_,6 @@
             this.gossips.add(source.getUUID(), GossipType.MINOR_NEGATIVE, 25);
         } else if (type == ReputationEventType.VILLAGER_KILLED) {
             this.gossips.add(source.getUUID(), GossipType.MAJOR_NEGATIVE, 25);
-        }
-
-        Player tradingPlayer = this.getTradingPlayer();
-        if (tradingPlayer != null && tradingPlayer.getUUID().equals(source.getUUID())) {
-            this.updateSpecialPrices(tradingPlayer);
-            this.resendOffersToTradingPlayer();
         }
     }
 
