--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -90,8 +_,11 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.timeline.Timelines;
+import org.bukkit.Bukkit;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
+import org.bukkit.event.entity.EntityTransformEvent;
+import org.bukkit.event.entity.VillagerReplenishTradeEvent;
 
 public class Villager extends AbstractVillager implements VillagerDataHolder, ReputationEventHandler {
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -118,7 +_,7 @@
     private boolean increaseProfessionLevelOnUpdate;
     private @Nullable Player lastTradedPlayer;
     private int foodLevel = 0;
-    private final GossipContainer gossips = new GossipContainer();
+    private final GossipContainer gossips = new GossipContainer(this); // CraftBukkit - add constructor parameter in Reputation
     private long lastGossipTime;
     private long lastGossipDecayTime = 0L;
     private int villagerXp = 0;
@@ -180,6 +_,8 @@
         MemoryModuleType.MEETING_POINT,
         (villager, poiType) -> poiType.is(PoiTypes.MEETING)
     );
+    private boolean isLobotomized = false; public boolean isLobotomized() { return this.isLobotomized; } // Purpur - Lobotomize stuck villagers
+    private int notLobotomizedCount = 0; // Purpur - Lobotomize stuck villagers
 
     public Villager(EntityType<? extends Villager> entityType, Level level) {
         super(entityType, level);
@@ -189,6 +_,110 @@
         this.setCanPickUpLoot(true);
     }
 
+    // Purpur start - Allow leashing villagers
+    @Override
+    public boolean canBeLeashed() {
+        return level().purpurConfig.villagerCanBeLeashed;
+    }
+    // Purpur end - Allow leashing villagers
+
+    // Purpur start - Lobotomize stuck villagers
+    private boolean checkLobotomized() {
+        int interval = this.level().purpurConfig.villagerLobotomizeCheckInterval;
+        boolean shouldCheckForTradeLocked = this.level().purpurConfig.villagerLobotomizeWaitUntilTradeLocked;
+        if (this.notLobotomizedCount > 3) {
+            // check half as often if not lobotomized for the last 3+ consecutive checks
+            interval *= 2;
+        }
+        if (this.level().getGameTime() % interval == 0) {
+            // offset Y for short blocks like dirt_path/farmland
+            this.isLobotomized = !(shouldCheckForTradeLocked && this.getVillagerXp() == 0) && !canTravelFrom(net.minecraft.core.BlockPos.containing(this.position().x, this.getBoundingBox().minY + 0.0625D, this.position().z));
+
+            if (this.isLobotomized) {
+                this.notLobotomizedCount = 0;
+            } else {
+                this.notLobotomizedCount++;
+            }
+        }
+        return this.isLobotomized;
+    }
+
+    private boolean canTravelFrom(net.minecraft.core.BlockPos pos) {
+        return canTravelTo(pos.east()) || canTravelTo(pos.west()) || canTravelTo(pos.north()) || canTravelTo(pos.south());
+    }
+
+    private boolean canTravelTo(net.minecraft.core.BlockPos pos) {
+        net.minecraft.world.level.block.state.BlockState state = this.level().getBlockStateIfLoaded(pos);
+        if (state == null) {
+            // chunk not loaded
+            return false;
+        }
+        net.minecraft.world.level.block.Block bottom = state.getBlock();
+        if (bottom instanceof net.minecraft.world.level.block.FenceBlock ||
+                bottom instanceof net.minecraft.world.level.block.FenceGateBlock ||
+                bottom instanceof net.minecraft.world.level.block.WallBlock) {
+            // bottom block is too tall to get over
+            return false;
+        }
+        net.minecraft.world.level.block.Block top = level().getBlockState(pos.above()).getBlock();
+        // only if both blocks have no collision
+        return !bottom.hasCollision && !top.hasCollision;
+    }
+    // Purpur end - Lobotomize stuck villagers
+
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.villagerRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.villagerRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.villagerControllable;
+    }
+
+    @Override
+    protected void registerGoals() {
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this));
+        if (level().purpurConfig.villagerFollowEmeraldBlock) this.goalSelector.addGoal(3, new net.minecraft.world.entity.ai.goal.TemptGoal(this, 1.0D, TEMPT_ITEMS, false)); // Purpur - Villagers follow emerald blocks
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.villagerMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.villagerScale);
+        this.getAttribute(Attributes.TEMPT_RANGE).setBaseValue(this.level().purpurConfig.villagerTemptRange); // Purpur - Villagers follow emerald blocks
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.villagerTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.villagerAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
+    // Purpur start - Check mobGriefing Overrides
+    @Override
+    protected Boolean checkEntityPickUpLootOverride() {
+        return this.level().purpurConfig.villagerCanPickUpLoot;
+    }
+    // Purpur end - Check mobGriefing Overrides
+
     @Override
     public Brain<Villager> getBrain() {
         return (Brain<Villager>) super.getBrain();
@@ -234,11 +_,35 @@
         return this.assignProfessionWhenSpawned;
     }
 
+    // Paper start - EAR 2
+    @Override
+    public void inactiveTick() {
+        // SPIGOT-3874, SPIGOT-3894, SPIGOT-3846, SPIGOT-5286 :(
+        if (this.getUnhappyCounter() > 0) {
+            this.setUnhappyCounter(this.getUnhappyCounter() - 1);
+        }
+        if (this.aware && this.isEffectiveAi()) { // Paper - Do not tick AI for inactive unaware mobs
+            if (this.level().spigotConfig.tickInactiveVillagers) {
+                this.customServerAiStep(this.level().getMinecraftWorld());
+            } else {
+                this.customServerAiStep(this.level().getMinecraftWorld(), true);
+            }
+        }
+        this.maybeDecayGossip();
+        super.inactiveTick();
+    }
+    // Paper end - EAR 2
+
     @Override
     protected void customServerAiStep(ServerLevel level) {
+        // Paper start - EAR 2
+        this.customServerAiStep(level, false);
+    }
+    protected void customServerAiStep(ServerLevel level, final boolean inactive) {
+        // Paper end - EAR 2
         ProfilerFiller profiler = Profiler.get();
         profiler.push("villagerBrain");
-        this.getBrain().tick(level, this);
+        if (!inactive) this.getBrain().tick(level, this); // Paper - EAR 2
         profiler.pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
@@ -252,7 +_,7 @@
                     this.increaseProfessionLevelOnUpdate = false;
                 }
 
-                this.addEffect(new MobEffectInstance(MobEffects.REGENERATION, 200, 0));
+                this.addEffect(new MobEffectInstance(MobEffects.REGENERATION, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.VILLAGER_TRADE); // CraftBukkit
             }
         }
 
@@ -262,7 +_,7 @@
             this.lastTradedPlayer = null;
         }
 
-        if (!this.isNoAi() && this.random.nextInt(100) == 0) {
+        if (!inactive && !this.isNoAi() && this.random.nextInt(100) == 0) { // Paper - EAR 2
             Raid raid = level.getRaidAt(this.blockPosition());
             if (raid != null && raid.isActive() && !raid.isOver()) {
                 level.broadcastEntityEvent(this, (byte)42);
@@ -273,6 +_,7 @@
             this.stopTrading();
         }
 
+        if (inactive) return; // Paper - EAR 2
         super.customServerAiStep(level);
     }
 
@@ -289,7 +_,7 @@
     @Override
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping()) {
+        if (itemStack.is(Items.VILLAGER_SPAWN_EGG) || !this.isAlive() || this.isTrading() || this.isSleeping() || player.isSecondaryUseActive()) {
             return super.mobInteract(player, hand);
         }
 
@@ -368,7 +_,13 @@
         this.updateDemand();
 
         for (MerchantOffer offer : this.getOffers()) {
-            offer.resetUses();
+            // CraftBukkit start
+            VillagerReplenishTradeEvent event = new VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), offer.asBukkit());
+            Bukkit.getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+                offer.resetUses();
+            }
+            // CraftBukkit end
         }
 
         this.resendOffersToTradingPlayer();
@@ -429,7 +_,13 @@
         int missedUpdates = 2 - this.numberOfRestocksToday;
         if (missedUpdates > 0) {
             for (MerchantOffer offer : this.getOffers()) {
-                offer.resetUses();
+                // CraftBukkit start
+                VillagerReplenishTradeEvent event = new VillagerReplenishTradeEvent((org.bukkit.entity.Villager) this.getBukkitEntity(), offer.asBukkit());
+                Bukkit.getPluginManager().callEvent(event);
+                if (!event.isCancelled()) {
+                    offer.resetUses();
+                }
+                // CraftBukkit end
             }
         }
 
@@ -450,6 +_,7 @@
         int reputation = this.getPlayerReputation(player);
         if (reputation != 0) {
             for (MerchantOffer offer : this.getOffers()) {
+                if (offer.ignoreDiscounts) continue; // Paper - Add ignore discounts API
                 offer.addToSpecialPriceDiff(-Mth.floor(reputation * offer.getPriceMultiplier()));
             }
         }
@@ -459,6 +_,7 @@
             int amplifier = effect.getAmplifier();
 
             for (MerchantOffer offer : this.getOffers()) {
+                if (offer.ignoreDiscounts) continue; // Paper - Add ignore discounts API
                 double modifier = 0.3 + 0.0625 * amplifier;
                 int costReduction = (int)Math.floor(modifier * offer.getBaseCostA().getCount());
                 offer.addToSpecialPriceDiff(-Math.max(costReduction, 1));
@@ -584,7 +_,7 @@
         }
 
         if (offer.shouldRewardExp()) {
-            this.level().addFreshEntity(new ExperienceOrb(this.level(), this.getX(), this.getY() + 0.5, this.getZ(), popXp));
+            this.level().addFreshEntity(new ExperienceOrb(this.level(), this.getX(), this.getY() + 0.5, this.getZ(), popXp, org.bukkit.entity.ExperienceOrb.SpawnReason.VILLAGER_TRADE, this.getTradingPlayer(), this)); // Paper
         }
     }
 
@@ -602,7 +_,7 @@
 
     @Override
     public void die(DamageSource source) {
-        LOGGER.info("Villager {} died, message: '{}'", this, source.getLocalizedDeathMessage(this).getString());
+        if (org.spigotmc.SpigotConfig.logVillagerDeaths) LOGGER.info("Villager {} died, message: '{}'", this, source.getLocalizedDeathMessage(this).getString()); // Spigot
         Entity murderer = source.getEntity();
         if (murderer != null) {
             this.tellWitnessesThatIWasMurdered(murderer);
@@ -701,7 +_,7 @@
         return VillagerData.canLevelUp(currentLevel) && this.villagerXp >= VillagerData.getMaxXpPerLevel(currentLevel);
     }
 
-    private void increaseMerchantCareer(ServerLevel level) {
+    public void increaseMerchantCareer(ServerLevel level) {
         this.setVillagerData(this.getVillagerData().withLevel(this.getVillagerData().level() + 1));
         this.updateTrades(level);
     }
@@ -761,13 +_,21 @@
 
     @Override
     public void thunderHit(ServerLevel level, LightningBolt lightningBolt) {
-        if (level.getDifficulty() != Difficulty.PEACEFUL) {
-            LOGGER.info("Villager {} was struck by lightning {}.", this, lightningBolt);
+        if (level.getDifficulty() != Difficulty.PEACEFUL && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityTypes.WITCH, (timer) -> {})) {
+            // Paper - Add EntityZapEvent; move log down, event can cancel
             Witch witch = this.convertTo(EntityTypes.WITCH, ConversionParams.single(this, false, false), w -> {
+                // Paper start - Add EntityZapEvent
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityZapEvent(this, lightningBolt, w).isCancelled()) {
+                    return false;
+                }
+                if (org.spigotmc.SpigotConfig.logVillagerDeaths) LOGGER.info("Villager {} was struck by lightning {}.", this, lightningBolt); // Move down
+                // Paper end - Add EntityZapEvent
                 w.finalizeSpawn(level, level.getCurrentDifficultyAt(w.blockPosition()), EntitySpawnReason.CONVERSION, null);
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, w);
                 w.setPersistenceRequired();
                 this.releaseAllPois();
-            });
+                return true; // Paper - Add EntityZapEvent
+            }, EntityTransformEvent.TransformReason.LIGHTNING, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING); // CraftBukkit
             if (witch == null) {
                 super.thunderHit(level, lightningBolt);
             }
@@ -784,7 +_,10 @@
     @Override
     public boolean wantsToPickUp(ServerLevel level, ItemStack itemStack) {
         Item item = itemStack.getItem();
-        return (itemStack.is(ItemTags.VILLAGER_PICKS_UP) || this.getVillagerData().profession().value().requestedItems().contains(item))
+        // Neo: Allow Farmer Villagers to pick up SpecialPlantable items.
+        boolean isFarmerDesiredSeed = (itemStack.getItem() instanceof net.neoforged.neoforge.common.SpecialPlantable specialPlantable && specialPlantable.villagerCanPlantItem(this))
+                && this.getVillagerData().profession().value().secondaryPoi().stream().anyMatch(secondaryPoi -> secondaryPoi.defaultBlockState().is(net.neoforged.neoforge.common.Tags.Blocks.VILLAGER_FARMLANDS));
+        return (itemStack.is(ItemTags.VILLAGER_PICKS_UP) || this.getVillagerData().profession().value().requestedItems().contains(item) || isFarmerDesiredSeed)
             && this.getInventory().canAddItem(itemStack);
     }
 
@@ -806,16 +_,24 @@
     }
 
     @Override
-    protected void updateTrades(ServerLevel level) {
+    public void updateTrades(ServerLevel level) {
+        // Paper start - More vanilla friendly methods to update trades
+        this.updateTrades(level, -1);
+    }
+
+    public boolean updateTrades(final ServerLevel level, int amount) {
+        // Paper end - More vanilla friendly methods to update trades
         VillagerData data = this.getVillagerData();
         VillagerProfession profession = data.profession().value();
         ResourceKey<TradeSet> trades = profession.getTrades(data.level());
         if (trades != null) {
-            this.addOffersFromTradeSet(level, this.getOffers(), trades);
+            this.addOffersFromTradeSet(level, this.getOffers(), trades, amount); // Paper - More vanilla friendly methods to update trades
             if (SharedConstants.DEBUG_UNLOCK_ALL_TRADES && data.level() < 5) {
                 this.increaseMerchantCareer(level);
             }
+            return true; // Paper - More vanilla friendly methods to update trades
         }
+        return false; // Paper - More vanilla friendly methods to update trades
     }
 
     public void gossip(ServerLevel level, Villager target, long timestamp) {
@@ -847,7 +_,7 @@
                 .limit(5L)
                 .toList();
             if (nearbyVillagersThatWantAGolem.size() >= villagersNeededToAgree) {
-                if (!SpawnUtil.trySpawnMob(
+                if (!SpawnUtil.trySpawnMob( // Paper - Set Golem Last Seen to stop it from spawning another one - switch to isPresent
                         EntityTypes.IRON_GOLEM,
                         EntitySpawnReason.MOB_SUMMONED,
                         level,
@@ -856,9 +_,11 @@
                         8,
                         6,
                         SpawnUtil.Strategy.LEGACY_IRON_GOLEM,
-                        false
+                        false,
+                        org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.VILLAGE_DEFENSE, // CraftBukkit,
+                        () -> {GolemSensor.golemDetected(this);} // Paper - Set Golem Last Seen to stop it from spawning another one
                     )
-                    .isEmpty()) {
+                    .isPresent()) { // Paper - Set Golem Last Seen to stop it from spawning another one - switch to isPresent
                     nearbyVillagers.forEach(GolemSensor::golemDetected);
                 }
             }
