--- a/net/minecraft/world/entity/animal/wolf/WolfVariant.java
+++ b/net/minecraft/world/entity/animal/wolf/WolfVariant.java
@@ -6,10 +_,10 @@
 import net.minecraft.core.ClientAsset;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.world.entity.variant.PriorityProvider;
 import net.minecraft.world.entity.variant.SpawnCondition;
 import net.minecraft.world.entity.variant.SpawnContext;
@@ -32,7 +_,7 @@
             )
             .apply(i, WolfVariant::new)
     );
-    public static final Codec<Holder<WolfVariant>> CODEC = RegistryCodecs.holder(Registries.WOLF_VARIANT);
+    public static final Codec<Holder<WolfVariant>> CODEC = RegistryFixedCodec.create(Registries.WOLF_VARIANT);
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<WolfVariant>> STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.WOLF_VARIANT);
 
     private WolfVariant(WolfVariant.AssetInfo adultInfo, WolfVariant.AssetInfo babyInfo) {
