--- a/net/minecraft/world/entity/animal/wolf/WolfSoundVariant.java
+++ b/net/minecraft/world/entity/animal/wolf/WolfSoundVariant.java
@@ -4,16 +_,16 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.sounds.SoundEvent;
 
 public record WolfSoundVariant(WolfSoundVariant.WolfSoundSet adultSounds, WolfSoundVariant.WolfSoundSet babySounds) {
     public static final Codec<WolfSoundVariant> DIRECT_CODEC = getWolfSoundVariantCodec();
     public static final Codec<WolfSoundVariant> NETWORK_CODEC = getWolfSoundVariantCodec();
-    public static final Codec<Holder<WolfSoundVariant>> CODEC = RegistryCodecs.holder(Registries.WOLF_SOUND_VARIANT);
+    public static final Codec<Holder<WolfSoundVariant>> CODEC = RegistryFixedCodec.create(Registries.WOLF_SOUND_VARIANT);
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<WolfSoundVariant>> STREAM_CODEC = ByteBufCodecs.holderRegistry(
         Registries.WOLF_SOUND_VARIANT
     );
