--- a/net/minecraft/world/entity/animal/wolf/Wolf.java
+++ b/net/minecraft/world/entity/animal/wolf/Wolf.java
@@ -272,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         if (this.isAngry()) {
             return this.getSoundVariant().value().growlSound().value();
         } else if (this.random.nextInt(3) == 0) {
@@ -285,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_406243_) {
+    public SoundEvent getHurtSound(DamageSource p_406243_) {
         return this.canArmorAbsorb(p_406243_) ? SoundEvents.WOLF_ARMOR_DAMAGE : this.getSoundVariant().value().hurtSound().value();
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return this.getSoundVariant().value().deathSound().value();
     }
 
     @Override
-    protected float getSoundVolume() {
+    public float getSoundVolume() {
         return 0.4F;
     }
 
@@ -459,10 +_,11 @@
         Item item = itemstack.getItem();
         if (this.isTame()) {
             if (this.isFood(itemstack) && this.getHealth() < this.getMaxHealth()) {
-                this.usePlayerItem(p_406380_, p_406261_, itemstack);
                 FoodProperties foodproperties = itemstack.get(DataComponents.FOOD);
                 float f = foodproperties != null ? foodproperties.nutrition() : 1.0F;
                 this.heal(2.0F * f);
+                this.usePlayerItem(p_406380_, p_406261_, itemstack);
+                this.gameEvent(GameEvent.EAT); // Neo: add EAT game event
                 return InteractionResult.SUCCESS;
             }
 
@@ -514,7 +_,7 @@
     }
 
     private void tryToTame(Player p_406358_) {
-        if (this.random.nextInt(3) == 0) {
+        if (this.random.nextInt(3) == 0  && !net.neoforged.neoforge.event.EventHooks.onAnimalTame(this, p_406358_)) {
             this.tame(p_406358_);
             this.navigation.stop();
             this.setTarget(null);
