--- a/net/minecraft/world/entity/animal/wolf/Wolf.java
+++ b/net/minecraft/world/entity/animal/wolf/Wolf.java
@@ -83,6 +_,11 @@
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec3;
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.event.entity.EntityTargetEvent;
+// CraftBukkit end
 import org.jspecify.annotations.Nullable;
 
 public class Wolf extends TamableAnimal implements NeutralMob {
@@ -395,27 +_,39 @@
         if (this.isInvulnerableTo(level, source)) {
             return false;
         } else {
+            // CraftBukkit start
+            boolean result = super.hurtServer(level, source, damage);
+            if (!result) {
+                return result;
+            }
+            // CraftBukkit end
             this.setOrderedToSit(false);
-            return super.hurtServer(level, source, damage);
+            return result; // CraftBukkit
         }
     }
 
     @Override
-    protected void actuallyHurt(ServerLevel level, DamageSource source, float damage) {
+    public boolean actuallyHurt(ServerLevel level, DamageSource source, float damage, EntityDamageEvent event) { // CraftBukkit - void -> boolean
         if (!this.canArmorAbsorb(source)) {
-            super.actuallyHurt(level, source, damage);
+            return super.actuallyHurt(level, source, damage, event); // CraftBukkit
         } else {
-            ItemStack armorBefore = this.getBodyArmorItem();
-            int damageBefore = armorBefore.getDamageValue();
-            int maxDamage = armorBefore.getMaxDamage();
-            armorBefore.hurtAndBreak(Mth.ceil(damage), this, EquipmentSlot.BODY);
-            if (Crackiness.WOLF_ARMOR.byDamage(damageBefore, maxDamage) != Crackiness.WOLF_ARMOR.byDamage(this.getBodyArmorItem())) {
+            // CraftBukkit start - SPIGOT-7815: if the damage was cancelled, no need to run the wolf armor behaviour
+            if (event.isCancelled()) {
+                return false;
+            }
+            // CraftBukkit end
+            ItemStack itemstack = this.getBodyArmorItem();
+            int i = itemstack.getDamageValue();
+            int j = itemstack.getMaxDamage();
+
+            itemstack.hurtAndBreak(Mth.ceil(damage), this, EquipmentSlot.BODY);
+            if (Crackiness.WOLF_ARMOR.byDamage(i, j) != Crackiness.WOLF_ARMOR.byDamage(this.getBodyArmorItem())) {
                 this.playSound(SoundEvents.WOLF_ARMOR_CRACK);
-                level.sendParticles(
-                    new ItemParticleOption(ParticleTypes.ITEM, Items.ARMADILLO_SCUTE), this.getX(), this.getY() + 1.0, this.getZ(), 20, 0.2, 0.1, 0.2, 0.1
-                );
+                level.sendParticles(new ItemParticleOption(ParticleTypes.ITEM, Items.ARMADILLO_SCUTE), this.getX(), this.getY() + 1.0D, this.getZ(), 20, 0.2D, 0.1D, 0.2D, 0.1D);
             }
+
         }
+        return false; // CraftBukkit
     }
 
     private boolean canArmorAbsorb(DamageSource source) {
@@ -426,7 +_,7 @@
     protected void applyTamingSideEffects() {
         if (this.isTame()) {
             this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(40.0);
-            this.setHealth(40.0F);
+            this.setHealth(this.getMaxHealth()); // CraftBukkit - 40.0 -> getMaxHealth()
         } else {
             this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(8.0);
         }
@@ -476,7 +_,7 @@
                     this.setOrderedToSit(!this.isOrderedToSit());
                     this.jumping = false;
                     this.navigation.stop();
-                    this.setTarget(null);
+                    this.setTarget((LivingEntity) null, EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit - reason
                     return InteractionResult.SUCCESS.withoutItem();
                 }
 
@@ -499,7 +_,8 @@
     }
 
     private void tryToTame(Player player) {
-        if (this.random.nextInt(3) == 0) {
+        // CraftBukkit - added event call and isCancelled check.
+        if (this.random.nextInt(3) == 0 && !CraftEventFactory.callEntityTameEvent(this, player).isCancelled()) {
             this.tame(player);
             this.navigation.stop();
             this.setTarget(null);
