--- a/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -78,7 +_,7 @@
             }
 
             HitResult hitresult = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity, this.getClipType());
-            if (hitresult.getType() != HitResult.Type.MISS) {
+            if (hitresult.getType() != HitResult.Type.MISS && !net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, hitresult)) {
                 this.hitTargetOrDeflectSelf(hitresult);
             }
 
@@ -185,7 +_,7 @@
         this.setDeltaMovement(vec3);
     }
 
-    public void assignDirectionalMovement(Vec3 p_347481_, double p_347589_) {
+    private void assignDirectionalMovement(Vec3 p_347481_, double p_347589_) {
         this.setDeltaMovement(p_347481_.normalize().scale(p_347589_));
         this.hasImpulse = true;
     }
