--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -4,6 +_,7 @@
 import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
 import java.util.Arrays;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
 import javax.annotation.Nullable;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -34,7 +_,6 @@
 import net.minecraft.world.entity.SlotAccess;
 import net.minecraft.world.entity.ai.attributes.Attributes;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
@@ -48,6 +_,7 @@
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.entity.CraftItem;
 
 public abstract class AbstractArrow extends Projectile {
     private static final double ARROW_BASE_DAMAGE = 2.0;
@@ -72,6 +_,18 @@
     @Nullable
     public ItemStack firedFromWeapon = null;
 
+    // Spigot Start
+    @Override
+    public void inactiveTick()
+    {
+        if ( this.inGround )
+        {
+            this.life += 1;
+        }
+        super.inactiveTick();
+    }
+    // Spigot End
+
     protected AbstractArrow(EntityType<? extends AbstractArrow> p_331098_, Level p_331626_) {
         super(p_331098_, p_331626_);
     }
@@ -85,7 +_,23 @@
         ItemStack p_309031_,
         @Nullable ItemStack p_345487_
     ) {
+        // CraftBukkit start - handle the owner before the rest of things
+        this(p_36717_, p_346045_, p_344855_, p_345999_, p_36719_, p_309031_, p_345487_, null);
+    }
+
+    protected AbstractArrow(
+        EntityType<? extends AbstractArrow> p_36717_,
+        double p_346045_,
+        double p_344855_,
+        double p_345999_,
+        Level p_36719_,
+        ItemStack p_309031_,
+        @Nullable ItemStack p_345487_,
+        @Nullable LivingEntity ownerEntity
+    ) {
         this(p_36717_, p_36719_);
+        this.setOwner(ownerEntity);
+        // CraftBukkit end
         this.pickupItemStack = p_309031_.copy();
         this.setCustomName(p_309031_.get(DataComponents.CUSTOM_NAME));
         Unit unit = p_309031_.remove(DataComponents.INTANGIBLE_PROJECTILE);
@@ -112,8 +_,7 @@
     protected AbstractArrow(
         EntityType<? extends AbstractArrow> p_36721_, LivingEntity p_345310_, Level p_36722_, ItemStack p_309145_, @Nullable ItemStack p_345000_
     ) {
-        this(p_36721_, p_345310_.getX(), p_345310_.getEyeY() - 0.1F, p_345310_.getZ(), p_36722_, p_309145_, p_345000_);
-        this.setOwner(p_345310_);
+        this(p_36721_, p_345310_.getX(), p_345310_.getEyeY() - 0.1F, p_345310_.getZ(), p_36722_, p_309145_, p_345000_, p_345310_);
     }
 
     public void setSoundEvent(SoundEvent p_36741_) {
@@ -188,7 +_,7 @@
             this.shakeTime--;
         }
 
-        if (this.isInWaterOrRain() || blockstate.is(Blocks.POWDER_SNOW)) {
+        if (this.isInWaterOrRain() || blockstate.is(Blocks.POWDER_SNOW) || this.isInFluidType((fluidType, height) -> this.canFluidExtinguish(fluidType))) {
             this.clearFire();
         }
 
@@ -201,6 +_,7 @@
 
             this.inGroundTime++;
         } else {
+            if (tickCount > 200) this.tickDespawn(); // Paper - tick despawnCounter regardless after 10 seconds
             this.inGroundTime = 0;
             Vec3 vec32 = this.position();
             Vec3 vec33 = vec32.add(vec3);
@@ -224,8 +_,10 @@
                     }
                 }
 
-                if (hitresult != null && !flag) {
-                    ProjectileDeflection projectiledeflection = this.hitTargetOrDeflectSelf(hitresult);
+                if (hitresult != null && hitresult.getType() != HitResult.Type.MISS && !flag) {
+                    if (net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, hitresult))
+                        break;
+                    ProjectileDeflection projectiledeflection = this.preHitTargetOrDeflectSelf(hitresult); // CraftBukkit - projectile hit event
                     this.hasImpulse = true;
                     if (projectiledeflection != ProjectileDeflection.NONE) {
                         break;
@@ -291,6 +_,19 @@
         }
     }
 
+    // Paper start - Fix cancelling ProjectileHitEvent for piercing arrows
+    @Override
+    public ProjectileDeflection preHitTargetOrDeflectSelf(HitResult hitResult) {
+        if (hitResult instanceof EntityHitResult entityHitResult && this.hitCancelled && this.getPierceLevel() > 0) {
+            if (this.piercingIgnoreEntityIds == null) {
+                this.piercingIgnoreEntityIds = new IntOpenHashSet(5);
+            }
+            this.piercingIgnoreEntityIds.add(entityHitResult.getEntity().getId());
+        }
+        return super.preHitTargetOrDeflectSelf(hitResult);
+    }
+    // Paper end - Fix cancelling ProjectileHitEvent for piercing arrows
+
     @Override
     protected double getDefaultGravity() {
         return 0.05;
@@ -319,7 +_,8 @@
 
     protected void tickDespawn() {
         this.life++;
-        if (this.life >= 1200) {
+        if (this.life >= (pickup == Pickup.CREATIVE_ONLY ? this.level().paperConfig().entities.spawning.creativeArrowDespawnRate.value() : (pickup == Pickup.DISALLOWED ? this.level().paperConfig().entities.spawning.nonPlayerArrowDespawnRate.value() : ((this instanceof ThrownTrident) ? this.level().spigotConfig.tridentDespawnRate : this.level().spigotConfig.arrowDespawnRate)))) { // Spigot // Paper - Configurable non-player arrow despawn rate; TODO: Extract this to init?
+            this.setRemoved$cause(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             this.discard();
         }
     }
@@ -357,7 +_,7 @@
             }
 
             if (this.piercingIgnoreEntityIds.size() >= this.getPierceLevel() + 1) {
-                this.discard();
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
                 return;
             }
 
@@ -372,11 +_,18 @@
         if (entity1 instanceof LivingEntity livingentity1) {
             livingentity1.setLastHurtMob(entity);
         }
-
+        if (this.isCritArrow()) damagesource = damagesource.critical(); // Paper - add critical damage API
         boolean flag = entity.getType() == EntityType.ENDERMAN;
         int i = entity.getRemainingFireTicks();
         if (this.isOnFire() && !flag) {
-            entity.igniteForSeconds(5.0F);
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityCombustByEntityEvent combustEvent = new org.bukkit.event.entity.EntityCombustByEntityEvent(this.getBukkitEntity(), entity.getBukkitEntity(), 5.0F);
+            org.bukkit.Bukkit.getPluginManager().callEvent(combustEvent);
+            if (!combustEvent.isCancelled()) {
+                entity.igniteForSeconds$callEvent = false;
+                entity.igniteForSeconds(combustEvent.getDuration());
+            }
+            // CraftBukkit end
         }
 
         if (entity.hurt(damagesource, (float)j)) {
@@ -414,7 +_,7 @@
 
             this.playSound(this.soundEvent, 1.0F, 1.2F / (this.random.nextFloat() * 0.2F + 0.9F));
             if (this.getPierceLevel() <= 0) {
-                this.discard();
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
             }
         } else {
             entity.setRemainingFireTicks(i);
@@ -425,7 +_,7 @@
                     this.spawnAtLocation(this.getPickupItem(), 0.1F);
                 }
 
-                this.discard();
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
             }
         }
     }
@@ -440,6 +_,7 @@
             double d1 = Math.max(0.0, 1.0 - p_346111_.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE));
             Vec3 vec3 = this.getDeltaMovement().multiply(1.0, 0.0, 1.0).normalize().scale(d0 * 0.6 * d1);
             if (vec3.lengthSqr() > 0.0) {
+                p_346111_.pushingEntity = this; // Paper - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
                 p_346111_.push(vec3.x, 0.1, vec3.z);
             }
         }
@@ -505,7 +_,7 @@
     }
 
     @Override
-    protected boolean canHitEntity(Entity p_36743_) {
+    public boolean canHitEntity(Entity p_36743_) {
         return super.canHitEntity(p_36743_) && (this.piercingIgnoreEntityIds == null || !this.piercingIgnoreEntityIds.contains(p_36743_.getId()));
     }
 
@@ -566,10 +_,12 @@
         }
     }
 
+    public AtomicBoolean setOwner$resetPickup = new AtomicBoolean(true);
+
     @Override
     public void setOwner(@Nullable Entity p_36770_) {
         super.setOwner(p_36770_);
-
+        if (!setOwner$resetPickup.getAndSet(true)) return; // Paper - Fix PickupStatus getting reset
         this.pickup = switch (p_36770_) {
             case Player player when this.pickup == AbstractArrow.Pickup.DISALLOWED -> AbstractArrow.Pickup.ALLOWED;
             case OminousItemSpawner ominousitemspawner -> AbstractArrow.Pickup.DISALLOWED;
@@ -577,12 +_,32 @@
         };
     }
 
+    public void setOwner(@Nullable Entity entity, boolean resetPickup) {
+        setOwner$resetPickup.set(resetPickup);
+        this.setOwner(entity);
+    }
+
     @Override
     public void playerTouch(Player p_36766_) {
         if (!this.level().isClientSide && (this.inGround || this.isNoPhysics()) && this.shakeTime <= 0) {
-            if (this.tryPickup(p_36766_)) {
+            // CraftBukkit start
+            ItemStack itemstack = this.getPickupItem();
+            if (this.pickup == Pickup.ALLOWED && !itemstack.isEmpty() && p_36766_.getInventory().canHold(itemstack) > 0) {
+                net.minecraft.world.entity.item.ItemEntity item = new net.minecraft.world.entity.item.ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), itemstack);
+                org.bukkit.event.player.PlayerPickupArrowEvent event = new org.bukkit.event.player.PlayerPickupArrowEvent((org.bukkit.entity.Player) p_36766_.getBukkitEntity(), new CraftItem(this.level().getCraftServer(), item), (org.bukkit.entity.AbstractArrow) this.getBukkitEntity());
+                // event.setCancelled(!entityhuman.canPickUpLoot); TODO
+                this.level().getCraftServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+                itemstack = item.getItem();
+            }
+
+            if ((this.pickup == AbstractArrow.Pickup.ALLOWED && p_36766_.getInventory().add(itemstack)) || (this.pickup == AbstractArrow.Pickup.CREATIVE_ONLY && p_36766_.getAbilities().instabuild)) {
+                // CraftBukkit end
                 p_36766_.take(this, 1);
-                this.discard();
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
             }
         }
     }
@@ -595,7 +_,7 @@
         };
     }
 
-    protected ItemStack getPickupItem() {
+    public ItemStack getPickupItem() {
         return this.pickupItemStack.copy();
     }
 
@@ -640,7 +_,7 @@
         }
     }
 
-    protected void setPickupItemStack(ItemStack p_331486_) {
+    public void setPickupItemStack(ItemStack p_331486_) {
         if (!p_331486_.isEmpty()) {
             this.pickupItemStack = p_331486_;
         } else {
