--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -18,7 +_,7 @@
 
 public class Arrow extends AbstractArrow {
     private static final int EXPOSED_POTION_DECAY_TIME = 600;
-    private static final int NO_EFFECT_COLOR = -1;
+    public static final int NO_EFFECT_COLOR = -1;
     private static final EntityDataAccessor<Integer> ID_EFFECT_COLOR = SynchedEntityData.defineId(Arrow.class, EntityDataSerializers.INT);
     private static final byte EVENT_POTION_PUFF = 0;
 
@@ -46,7 +_,7 @@
     }
 
     @Override
-    protected void setPickupItemStack(ItemStack p_331667_) {
+    public void setPickupItemStack(ItemStack p_331667_) {
         super.setPickupItemStack(p_331667_);
         this.updateColor();
     }
@@ -120,13 +_,14 @@
                         mobeffectinstance.isAmbient(),
                         mobeffectinstance.isVisible()
                     ),
-                    entity
+                    entity,
+                    org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW // CraftBukkit
                 );
             }
         }
 
         for (MobEffectInstance mobeffectinstance1 : potioncontents.customEffects()) {
-            p_36873_.addEffect(mobeffectinstance1, entity);
+            p_36873_.addEffect(mobeffectinstance1, entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
         }
     }
 
