--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -36,11 +_,11 @@
         this.updateColor();
     }
 
-    public PotionContents getPotionContents() {
+    private PotionContents getPotionContents() {
         return this.getPickupItemStackOrigin().getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY);
     }
 
-    public void setPotionContents(PotionContents p_331534_) {
+    private void setPotionContents(PotionContents p_331534_) {
         this.getPickupItemStackOrigin().set(DataComponents.POTION_CONTENTS, p_331534_);
         this.updateColor();
     }
@@ -51,7 +_,7 @@
         this.updateColor();
     }
 
-    public void updateColor() {
+    private void updateColor() {
         PotionContents potioncontents = this.getPotionContents();
         this.entityData.set(ID_EFFECT_COLOR, potioncontents.equals(PotionContents.EMPTY) ? -1 : potioncontents.getColor());
     }
