--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -120,13 +_,14 @@
                         mobeffectinstance.isAmbient(),
                         mobeffectinstance.isVisible()
                     ),
-                    entity
+                    entity,
+                    org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW // CraftBukkit
                 );
             }
         }
 
         for (MobEffectInstance mobeffectinstance1 : potioncontents.customEffects()) {
-            p_36873_.addEffect(mobeffectinstance1, entity);
+            p_36873_.addEffect(mobeffectinstance1, entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
         }
     }
 
