--- a/net/minecraft/world/entity/projectile/EvokerFangs.java
+++ b/net/minecraft/world/entity/projectile/EvokerFangs.java
@@ -6,7 +_,6 @@
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntityEvent;
 import net.minecraft.world.entity.EntityReference;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EntityTypes;
@@ -96,7 +_,7 @@
             }
 
             if (--this.lifeTicks < 0) {
-                this.discard();
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             }
         }
     }
@@ -105,7 +_,7 @@
         LivingEntity currentOwner = this.getOwner();
         if (entity.isAlive() && !entity.isInvulnerable() && entity != currentOwner) {
             if (currentOwner == null) {
-                entity.hurt(this.damageSources().magic(), 6.0F);
+                entity.hurt(this.damageSources().magic().eventEntityDamager(this), 6.0F); // CraftBukkit
             } else {
                 if (currentOwner.isAlliedTo(entity)) {
                     return;
@@ -120,7 +_,7 @@
     }
 
     @Override
-    public void handleEntityEvent(@EntityEvent.Value byte id) {
+    public void handleEntityEvent(byte id) {
         super.handleEntityEvent(id);
         if (id == 4) {
             this.clientSideAttackStarted = true;
