--- a/net/minecraft/world/entity/projectile/EyeOfEnder.java
+++ b/net/minecraft/world/entity/projectile/EyeOfEnder.java
@@ -63,6 +_,11 @@
     }
 
     public void signalTo(BlockPos p_36968_) {
+        // Paper start - Change EnderEye target without changing other things
+        this.signalTo(p_36968_, true);
+    }
+    public void signalTo(BlockPos p_36968_, boolean update) {
+        // Paper end - Change EnderEye target without changing other things
         double d0 = (double)p_36968_.getX();
         int i = p_36968_.getY();
         double d1 = (double)p_36968_.getZ();
@@ -79,8 +_,10 @@
             this.tz = d1;
         }
 
-        this.life = 0;
-        this.surviveAfterDeath = this.random.nextInt(5) > 0;
+        if (update) { // Paper - Change EnderEye target without changing other things
+            this.life = 0;
+            this.surviveAfterDeath = this.random.nextInt(5) > 0;
+        } // Paper - Change EnderEye target without changing other things
     }
 
     @Override
@@ -145,7 +_,7 @@
             this.life++;
             if (this.life > 80 && !this.level().isClientSide) {
                 this.playSound(SoundEvents.ENDER_EYE_DEATH, 1.0F, 1.0F);
-                this.discard();
+                this.discard(this.surviveAfterDeath ? org.bukkit.event.entity.EntityRemoveEvent.Cause.DROP : org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                 if (this.surviveAfterDeath) {
                     this.level().addFreshEntity(new ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), this.getItem()));
                 } else {
@@ -165,7 +_,12 @@
     @Override
     public void readAdditionalSaveData(CompoundTag p_36970_) {
         if (p_36970_.contains("Item", 10)) {
-            this.setItem(ItemStack.parse(this.registryAccess(), p_36970_.getCompound("Item")).orElse(this.getDefaultItem()));
+            // CraftBukkit start - SPIGOT-6103 summon, see also SPIGOT-5474
+            ItemStack itemstack = (ItemStack) ItemStack.parse(this.registryAccess(), p_36970_.getCompound("Item")).orElse(this.getDefaultItem());
+            if (!itemstack.isEmpty()) {
+                this.setItem(itemstack);
+            }
+            // CraftBukkit end
         } else {
             this.setItem(this.getDefaultItem());
         }
