--- a/net/minecraft/world/entity/projectile/EyeOfEnder.java
+++ b/net/minecraft/world/entity/projectile/EyeOfEnder.java
@@ -20,6 +_,10 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public class EyeOfEnder extends Entity implements ItemSupplier {
     private static final float MIN_CAMERA_DISTANCE_SQUARED = 12.25F;
     private static final float TOO_FAR_SIGNAL_HEIGHT = 8.0F;
@@ -72,6 +_,12 @@
     }
 
     public void signalTo(Vec3 target) {
+        // Paper start - Change EyeOfEnder target without changing other things
+        this.signalTo(target, true);
+    }
+
+    public void signalTo(final Vec3 target, final boolean update) {
+        // Paper end - Change EyeOfEnder target without changing other things
         Vec3 delta = target.subtract(this.position());
         double horizontalDistance = delta.horizontalDistance();
         if (horizontalDistance > 12.0) {
@@ -80,8 +_,10 @@
             this.target = target;
         }
 
-        this.life = 0;
-        this.surviveAfterDeath = this.random.nextInt(5) > 0;
+        if (update) { // Paper - Change EyeOfEnder target without changing other things
+            this.life = 0;
+            this.surviveAfterDeath = this.random.nextInt(5) > 0;
+        } // Paper - Change EyeOfEnder target without changing other things
     }
 
     @Override
@@ -102,7 +_,7 @@
             this.life++;
             if (this.life > 80 && !this.level().isClientSide()) {
                 this.playSound(SoundEvents.ENDER_EYE_DEATH, 1.0F, 1.0F);
-                this.discard();
+                this.discard(this.surviveAfterDeath ? EntityRemoveEvent.Cause.DROP : EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
                 if (this.surviveAfterDeath) {
                     this.level().addFreshEntity(new ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), this.getItem()));
                 } else {
