--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -29,12 +_,13 @@
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class FireworkRocketEntity extends Projectile implements ItemSupplier {
     public static final EntityDataAccessor<ItemStack> DATA_ID_FIREWORKS_ITEM = SynchedEntityData.defineId(
         FireworkRocketEntity.class, EntityDataSerializers.ITEM_STACK
     );
-    private static final EntityDataAccessor<OptionalInt> DATA_ATTACHED_TO_TARGET = SynchedEntityData.defineId(
+    public static final EntityDataAccessor<OptionalInt> DATA_ATTACHED_TO_TARGET = SynchedEntityData.defineId(
         FireworkRocketEntity.class, EntityDataSerializers.OPTIONAL_UNSIGNED_INT
     );
     public static final EntityDataAccessor<Boolean> DATA_SHOT_AT_ANGLE = SynchedEntityData.defineId(FireworkRocketEntity.class, EntityDataSerializers.BOOLEAN);
@@ -42,6 +_,7 @@
     public int lifetime;
     @Nullable
     public LivingEntity attachedToEntity;
+    @Nullable public java.util.UUID spawningEntity; // Paper
 
     public FireworkRocketEntity(EntityType<? extends FireworkRocketEntity> p_37027_, Level p_37028_) {
         super(p_37027_, p_37028_);
@@ -83,6 +_,21 @@
         this.setOwner(p_37052_);
     }
 
+    // Spigot Start - copied from tick
+    @Override
+    public void inactiveTick() {
+        this.life += 1;
+        if (!this.level().isClientSide && this.life > this.lifetime) {
+            // CraftBukkit start
+            if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+                this.explode();
+            }
+            // CraftBukkit end
+        }
+        super.inactiveTick();
+    }
+    // Spigot End
+
     @Override
     protected void defineSynchedData(SynchedEntityData.Builder p_326171_) {
         p_326171_.define(DATA_ID_FIREWORKS_ITEM, getDefaultItem());
@@ -149,7 +_,7 @@
 
         HitResult hitresult = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
         if (!this.noPhysics) {
-            this.hitTargetOrDeflectSelf(hitresult);
+            this.preHitTargetOrDeflectSelf(hitresult); // CraftBukkit - projectile hit event
             this.hasImpulse = true;
         }
 
@@ -173,7 +_,18 @@
         }
 
         if (!this.level().isClientSide && this.life > this.lifetime) {
-            this.explode();
+            // CraftBukkit start
+            if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+                this.explode();
+            }
+            // CraftBukkit end
+        }
+    }
+
+    @Override
+    protected void onHit(HitResult result) {
+        if (result.getType() == HitResult.Type.MISS || !net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, result)) {
+            super.onHit(result);
         }
     }
 
@@ -181,14 +_,18 @@
         this.level().broadcastEntityEvent(this, (byte)17);
         this.gameEvent(GameEvent.EXPLODE, this.getOwner());
         this.dealExplosionDamage();
-        this.discard();
+        this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit - add Bukkit remove cause
     }
 
     @Override
     protected void onHitEntity(EntityHitResult p_37071_) {
         super.onHitEntity(p_37071_);
         if (!this.level().isClientSide) {
-            this.explode();
+            // CraftBukkit start
+            if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+                this.explode();
+            }
+            // CraftBukkit end
         }
     }
 
@@ -197,7 +_,11 @@
         BlockPos blockpos = new BlockPos(p_37069_.getBlockPos());
         this.level().getBlockState(blockpos).entityInside(this.level(), blockpos, this);
         if (!this.level().isClientSide() && this.hasExplosion()) {
-            this.explode();
+            // CraftBukkit start
+            if (!CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+                this.explode();
+            }
+            // CraftBukkit end
         }
 
         super.onHitBlock(p_37069_);
@@ -269,6 +_,11 @@
         p_37075_.putInt("LifeTime", this.lifetime);
         p_37075_.put("FireworksItem", this.getItem().save(this.registryAccess()));
         p_37075_.putBoolean("ShotAtAngle", this.entityData.get(DATA_SHOT_AT_ANGLE));
+        // Paper start
+        if (this.spawningEntity != null) {
+            p_37075_.putUUID("SpawningEntity", this.spawningEntity);
+        }
+        // Paper end
     }
 
     @Override
@@ -289,6 +_,11 @@
         if (p_37073_.contains("ShotAtAngle")) {
             this.entityData.set(DATA_SHOT_AT_ANGLE, p_37073_.getBoolean("ShotAtAngle"));
         }
+        // Paper start
+        if (p_37073_.hasUUID("SpawningEntity")) {
+            this.spawningEntity = p_37073_.getUUID("SpawningEntity");
+        }
+        // Paper end
     }
 
     private List<FireworkExplosion> getExplosions() {
@@ -307,7 +_,7 @@
         return false;
     }
 
-    private static ItemStack getDefaultItem() {
+    public static ItemStack getDefaultItem() {
         return new ItemStack(Items.FIREWORK_ROCKET);
     }
 
