--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -27,17 +_,23 @@
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.projectiles.ProjectileSource;
 
 public abstract class Projectile extends Entity implements TraceableEntity {
     @Nullable
-    private UUID ownerUUID;
+    public UUID ownerUUID;
     @Nullable
-    private Entity cachedOwner;
-    private boolean leftOwner;
-    private boolean hasBeenShot;
+    public Entity cachedOwner;
+    public boolean leftOwner;
+    public boolean hasBeenShot;
     @Nullable
     private Entity lastDeflectedBy;
 
+    // CraftBukkit start
+    protected boolean hitCancelled = false;
+    // CraftBukkit end
+
     protected Projectile(EntityType<? extends Projectile> p_37248_, Level p_37249_) {
         super(p_37248_, p_37249_);
     }
@@ -47,15 +_,36 @@
             this.ownerUUID = p_37263_.getUUID();
             this.cachedOwner = p_37263_;
         }
-    }
+        // Paper start - Refresh ProjectileSource for projectiles
+        else {
+            this.ownerUUID = null;
+            this.cachedOwner = null;
+            this.projectileSource = null;
+        }
+        // Paper end - Refresh ProjectileSource for projectiles
+        this.refreshProjectileSource(false); // Paper
+    }
+
+    // Paper start - Refresh ProjectileSource for projectiles
+    public void refreshProjectileSource(boolean fillCache) {
+        if (fillCache) {
+            this.getOwner();
+        }
+        if (this.cachedOwner != null && !this.cachedOwner.isRemoved() && this.projectileSource == null && this.cachedOwner.getBukkitEntity() instanceof ProjectileSource projSource) {
+            this.projectileSource = projSource;
+        }
+    }
+    // Paper end - Refresh ProjectileSource for projectiles
 
     @Nullable
     @Override
     public Entity getOwner() {
         if (this.cachedOwner != null && !this.cachedOwner.isRemoved()) {
+            this.refreshProjectileSource(false); // Paper - Refresh ProjectileSource for projectiles
             return this.cachedOwner;
         } else if (this.ownerUUID != null && this.level() instanceof ServerLevel serverlevel) {
             this.cachedOwner = serverlevel.getEntity(this.ownerUUID);
+            this.refreshProjectileSource(false); // Paper - Refresh ProjectileSource for projectiles
             return this.cachedOwner;
         } else {
             return null;
@@ -88,6 +_,7 @@
         if (p_37262_.hasUUID("Owner")) {
             this.ownerUUID = p_37262_.getUUID("Owner");
             this.cachedOwner = null;
+            if (this instanceof ThrownEnderpearl && this.level() != null && this.level().paperConfig().fixes.disableUnloadedChunkEnderpearlExploit) { this.ownerUUID = null; } // Paper - Reset pearls when they stop being ticked; Don't store shooter name for pearls to block enderpearl travel exploit
         }
 
         this.leftOwner = p_37262_.getBoolean("LeftOwner");
@@ -162,8 +_,23 @@
         float f2 = Mth.cos(p_37254_ * (float) (Math.PI / 180.0)) * Mth.cos(p_37253_ * (float) (Math.PI / 180.0));
         this.shoot((double)f, (double)f1, (double)f2, p_37256_, p_37257_);
         Vec3 vec3 = p_37252_.getKnownMovement();
+        // Paper start - allow disabling relative velocity
+        if (!p_37252_.level().paperConfig().misc.disableRelativeProjectileVelocity) {
         this.setDeltaMovement(this.getDeltaMovement().add(vec3.x, p_37252_.onGround() ? 0.0 : vec3.y, vec3.z));
-    }
+        }
+        // Paper end - allow disabling relative velocity
+    }
+
+    // CraftBukkit start - call projectile hit event
+    public ProjectileDeflection preHitTargetOrDeflectSelf(HitResult movingobjectposition) {
+        org.bukkit.event.entity.ProjectileHitEvent event = CraftEventFactory.callProjectileHitEvent(this, movingobjectposition);
+        this.hitCancelled = event != null && event.isCancelled();
+        if (movingobjectposition.getType() == HitResult.Type.BLOCK || !this.hitCancelled) {
+            return this.hitTargetOrDeflectSelf(movingobjectposition);
+        }
+        return ProjectileDeflection.NONE;
+    }
+    // CraftBukkit end
 
     protected ProjectileDeflection hitTargetOrDeflectSelf(HitResult p_341949_) {
         if (p_341949_.getType() == HitResult.Type.ENTITY) {
@@ -186,7 +_,13 @@
     public boolean deflect(ProjectileDeflection p_341900_, @Nullable Entity p_341912_, @Nullable Entity p_341932_, boolean p_341948_) {
         if (!this.level().isClientSide) {
             p_341900_.deflect(this, p_341912_, this.random);
-            this.setOwner(p_341932_);
+            // Paper start - Fix PickupStatus getting reset
+            if (this instanceof AbstractArrow arrow) {
+                arrow.setOwner(p_341932_, false);
+            } else {
+                this.setOwner(p_341932_);
+            }
+            // Paper end - Fix PickupStatus getting reset
             this.onDeflection(p_341912_, p_341948_);
         }
 
@@ -219,6 +_,11 @@
     }
 
     protected void onHitBlock(BlockHitResult p_37258_) {
+        // CraftBukkit start - cancellable hit event
+        if (hitCancelled) {
+            return;
+        }
+        // CraftBukkit end
         BlockState blockstate = this.level().getBlockState(p_37258_.getBlockPos());
         blockstate.onProjectileHit(this.level(), blockstate, p_37258_, this);
     }
@@ -236,11 +_,20 @@
         }
     }
 
-    protected boolean canHitEntity(Entity p_37250_) {
+    public boolean canHitEntity(Entity p_37250_) {
         if (!p_37250_.canBeHitByProjectile()) {
             return false;
         } else {
             Entity entity = this.getOwner();
+            // Paper start - Cancel hit for vanished players
+            if (entity instanceof net.minecraft.server.level.ServerPlayer && entity instanceof net.minecraft.server.level.ServerPlayer) {
+                org.bukkit.entity.Player collided = (org.bukkit.entity.Player) entity.getBukkitEntity();
+                org.bukkit.entity.Player shooter = (org.bukkit.entity.Player) entity.getBukkitEntity();
+                if (!shooter.canSee(collided)) {
+                    return false;
+                }
+            }
+            // Paper end - Cancel hit for vanished players
             return entity == null || this.leftOwner || !entity.isPassengerOfSameVehicle(p_37250_);
         }
     }
@@ -284,7 +_,7 @@
         Entity entity = this.getOwner();
         return entity instanceof Player
             ? entity.mayInteract(p_150167_, p_150168_)
-            : entity == null || p_150167_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+            : entity == null || net.neoforged.neoforge.event.EventHooks.canEntityGrief(p_150167_, entity);
     }
 
     public boolean mayBreak(Level p_307481_) {
