--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -20,11 +_,11 @@
 import net.minecraft.world.entity.EntitySelector;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
-import net.minecraft.world.entity.MoveSimulationType;
 import net.minecraft.world.entity.TraceableEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.arrow.AbstractArrow;
 import net.minecraft.world.item.AdventureModePredicate;
+import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.level.GameType;
@@ -45,16 +_,13 @@
 public abstract class Projectile extends Entity implements TraceableEntity {
     private static final boolean DEFAULT_LEFT_OWNER = false;
     private static final boolean DEFAULT_HAS_BEEN_SHOT = false;
-    private static final String TAG_OWNER = "Owner";
-    private static final String TAG_LEFT_OWNER = "LeftOwner";
-    private static final String TAG_HAS_BEEN_SHOT = "HasBeenShot";
-    private static final String TAG_CAN_BREAK = "can_break";
-    public @Nullable EntityReference<Entity> owner;
+    protected @Nullable EntityReference<Entity> owner;
     private @Nullable AdventureModePredicate canBreak;
     public boolean leftOwner = false;
     private boolean leftOwnerChecked;
     public boolean hasBeenShot = false;
     private @Nullable Entity lastDeflectedBy;
+    protected boolean hitCancelled = false; // CraftBukkit
 
     protected Projectile(EntityType<? extends Projectile> type, Level level) {
         super(type, level);
@@ -101,9 +_,6 @@
         }
 
         output.putBoolean("HasBeenShot", this.hasBeenShot);
-        if (this.canBreak != null) {
-            output.store("can_break", AdventureModePredicate.CODEC, this.canBreak);
-        }
     }
 
     protected boolean ownedBy(Entity entity) {
@@ -113,9 +_,11 @@
     @Override
     protected void readAdditionalSaveData(ValueInput input) {
         this.setOwner(EntityReference.read(input, "Owner"));
+        // Paper start - Reset pearls when they stop being ticked; Don't store shooter name for pearls to block enderpearl travel exploit
+        if (this instanceof net.minecraft.world.entity.projectile.throwableitemprojectile.ThrownEnderpearl && this.level().paperConfig().fixes.disableUnloadedChunkEnderpearlExploit && this.level().paperConfig().misc.legacyEnderPearlBehavior) { this.owner = null; }
+        // Paper end - Reset pearls when they stop being ticked; Don't store shooter name for pearls to block enderpearl travel exploit
         this.leftOwner = input.getBooleanOr("LeftOwner", false);
         this.hasBeenShot = input.getBooleanOr("HasBeenShot", false);
-        this.canBreak = input.read("can_break", AdventureModePredicate.CODEC).orElse(null);
     }
 
     @Override
@@ -186,7 +_,14 @@
         float zd = Mth.cos(yRot * (float) (Math.PI / 180.0)) * Mth.cos(xRot * (float) (Math.PI / 180.0));
         this.shoot(xd, yd, zd, pow, uncertainty);
         Vec3 sourceMovement = source.getKnownMovement();
+        // Paper start - allow disabling relative velocity
+        if (!sourceMovement.isFinite()) {
+            sourceMovement = Vec3.ZERO;
+        }
+        if (!source.level().paperConfig().misc.disableRelativeProjectileVelocity) {
         this.setDeltaMovement(this.getDeltaMovement().add(sourceMovement.x, source.onGround() ? 0.0 : sourceMovement.y, sourceMovement.z));
+        }
+        // Paper end - allow disabling relative velocity
     }
 
     @Override
@@ -203,22 +_,39 @@
         this.resetFallDistance();
     }
 
-    public boolean canBreakBlockInAdventureMode(ServerLevel level, BlockInWorld blockInWorld) {
-        Entity owner = this.getOwner();
-        GameType worldGameMode = level.getServer().getDefaultGameType();
-        return (owner != null || worldGameMode == GameType.ADVENTURE) && !(owner instanceof Player player && player.gameMode() != GameType.ADVENTURE)
-            ? this.canBreak != null && this.canBreak.test(blockInWorld)
-            : true;
+    public boolean canBreakBlockInAdventureMode(BlockInWorld blockInWorld) {
+        return this.getOwner() instanceof Player player && player.gameMode() != GameType.ADVENTURE
+            ? true
+            : this.canBreak != null && this.canBreak.test(blockInWorld);
     }
 
     public static <T extends Projectile> T spawnProjectileFromRotation(
-        Projectile.ProjectileFactory<T> creator, ServerLevel serverLevel, ItemStack itemStack, LivingEntity source, float yOffset, float pow, float uncertainty
-    ) {
-        return spawnProjectile(
-            creator.create(serverLevel, source, itemStack),
-            serverLevel,
-            itemStack,
-            projectile -> projectile.shootFromRotation(source, source.getXRot(), source.getYRot(), yOffset, pow, uncertainty)
+            final Projectile.ProjectileFactory<T> creator,
+            final ServerLevel serverLevel,
+            final ItemStack itemStack,
+            final LivingEntity source,
+            final float yOffset,
+            final float pow,
+            final float uncertainty
+    ) {
+        // Paper start - PlayerLaunchProjectileEvent
+        return spawnProjectileFromRotationDelayed(creator, serverLevel, itemStack, source, yOffset, pow, uncertainty).spawn();
+    }
+    public static <T extends Projectile> Delayed<T> spawnProjectileFromRotationDelayed(
+            final Projectile.ProjectileFactory<T> creator,
+            final ServerLevel serverLevel,
+            final ItemStack itemStack,
+            final LivingEntity source,
+            final float yOffset,
+            final float pow,
+            final float uncertainty
+    ) {
+        return spawnProjectileDelayed(
+                // Paper end - PlayerLaunchProjectileEvent
+                creator.create(serverLevel, source, itemStack),
+                serverLevel,
+                itemStack,
+                projectile -> projectile.shootFromRotation(source, source.getXRot(), source.getYRot(), yOffset, pow, uncertainty)
         );
     }
 
@@ -239,21 +_,80 @@
     }
 
     public static <T extends Projectile> T spawnProjectileUsingShoot(
-        T projectile, ServerLevel serverLevel, ItemStack itemStack, double targetX, double targetY, double targetZ, float pow, float uncertainty
-    ) {
-        return spawnProjectile(projectile, serverLevel, itemStack, i -> projectile.shoot(targetX, targetY, targetZ, pow, uncertainty));
-    }
-
-    public static <T extends Projectile> T spawnProjectile(T projectile, ServerLevel serverLevel, ItemStack itemStack) {
+            final T projectile,
+            final ServerLevel serverLevel,
+            final ItemStack itemStack,
+            final double targetX,
+            final double targetY,
+            final double targetZ,
+            final float pow,
+            final float uncertainty
+    ) {
+        // Paper start - fixes and addition to spawn reason API
+        return spawnProjectileUsingShootDelayed(projectile, serverLevel, itemStack, targetX, targetY, targetZ, pow, uncertainty).spawn();
+    }
+
+    public static <T extends Projectile> Delayed<T> spawnProjectileUsingShootDelayed(
+            final T projectile,
+            final ServerLevel serverLevel,
+            final ItemStack itemStack,
+            final double targetX,
+            final double targetY,
+            final double targetZ,
+            final float pow,
+            final float uncertainty
+    ) {
+        return spawnProjectileDelayed(projectile, serverLevel, itemStack, i -> projectile.shoot(targetX, targetY, targetZ, pow, uncertainty));
+        // Paper end - fixes and addition to spawn reason API
+    }
+
+    public static <T extends Projectile> T spawnProjectile(final T projectile, final ServerLevel serverLevel, final ItemStack itemStack) {
         return spawnProjectile(projectile, serverLevel, itemStack, ignored -> {});
     }
 
-    public static <T extends Projectile> T spawnProjectile(T projectile, ServerLevel serverLevel, ItemStack itemStack, Consumer<T> shootFunction) {
-        shootFunction.accept(projectile);
-        serverLevel.addFreshEntity(projectile);
-        projectile.applyOnProjectileSpawned(serverLevel, itemStack);
-        return projectile;
-    }
+    public static <T extends Projectile> T spawnProjectile(
+            final T projectile, final ServerLevel serverLevel, final ItemStack itemStack, final Consumer<T> shootFunction
+    ) {
+        // Paper start - delayed projectile spawning
+        return spawnProjectileDelayed(projectile, serverLevel, itemStack, shootFunction).spawn();
+    }
+
+    public static <T extends Projectile> Delayed<T> spawnProjectileDelayed(T projectile, ServerLevel serverLevel, ItemStack itemStack, Consumer<T> adapter) {
+        // Paper end - delayed projectile spawning
+        adapter.accept(projectile);
+        return new Delayed<>(projectile, serverLevel, itemStack); // Paper - delayed projectile spawning
+    }
+
+    // Paper start - delayed projectile spawning
+    public record Delayed<T extends Projectile>(
+            T projectile,
+            ServerLevel world,
+            ItemStack projectileStack
+    ) {
+        // Taken from net.minecraft.world.entity.projectile.Projectile.spawnProjectile(T, net.minecraft.server.level.ServerLevel, net.minecraft.world.item.ItemStack, java.util.function.Consumer<T>)
+        public boolean attemptSpawn() {
+            if (!this.world.addFreshEntity(this.projectile)) return false;
+            this.projectile.applyOnProjectileSpawned(this.world, this.projectileStack);
+            return true;
+        }
+
+        public T spawn() {
+            this.attemptSpawn();
+            return this.projectile();
+        }
+
+        public boolean attemptSpawn(final org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+            if (!this.world.addFreshEntity(this.projectile, reason)) return false;
+            this.projectile.applyOnProjectileSpawned(this.world, this.projectileStack);
+            return true;
+        }
+
+        public T spawn(final org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+            this.attemptSpawn(reason);
+            return this.projectile();
+        }
+    }
+    // Paper end - delayed projectile spawning
 
     public void applyOnProjectileSpawned(ServerLevel serverLevel, ItemStack pickupItemStack) {
         EnchantmentHelper.onProjectileSpawned(serverLevel, pickupItemStack, this, item -> {});
@@ -265,28 +_,34 @@
         }
     }
 
+    // CraftBukkit start - call projectile hit event
+    public ProjectileDeflection preHitTargetOrDeflectSelf(HitResult hitresult) {
+        org.bukkit.event.entity.ProjectileHitEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callProjectileHitEvent(this, hitresult);
+        this.hitCancelled = event != null && event.isCancelled();
+        if (hitresult.getType() == HitResult.Type.BLOCK || !this.hitCancelled) {
+            return this.hitTargetOrDeflectSelf(hitresult);
+        }
+        return ProjectileDeflection.NONE;
+    }
+    // CraftBukkit end
+
     protected ProjectileDeflection hitTargetOrDeflectSelf(HitResult hitResult) {
-        if (hitResult instanceof EntityHitResult entityHitResult) {
+        if (hitResult.getType() == HitResult.Type.ENTITY) {
+            EntityHitResult entityHitResult = (EntityHitResult)hitResult;
             Entity entity = entityHitResult.getEntity();
             ProjectileDeflection deflection = entity.deflection(this);
             if (deflection != ProjectileDeflection.NONE) {
-                if (entity != this.lastDeflectedBy && this.deflect(deflection, entity, this.owner, false, 1.0)) {
+                if (entity != this.lastDeflectedBy && this.deflect(deflection, entity, this.owner, false)) {
                     this.lastDeflectedBy = entity;
                 }
 
                 return deflection;
             }
-        } else if (hitResult instanceof BlockHitResult blockHitResult) {
-            if (!blockHitResult.isWorldBorderHit()) {
-                BlockState collidedWith = this.level().getBlockState(blockHitResult.getBlockPos());
-                if (collidedWith.isAir()) {
-                    return ProjectileDeflection.NONE;
-                }
-            } else if (this.shouldBounceOnWorldBorder()) {
-                ProjectileDeflection deflection = ProjectileDeflection.REVERSE;
-                if (this.deflect(deflection, null, this.owner, false, 0.2)) {
-                    return deflection;
-                }
+        } else if (this.shouldBounceOnWorldBorder() && hitResult instanceof BlockHitResult blockHit && blockHit.isWorldBorderHit()) {
+            ProjectileDeflection deflection = ProjectileDeflection.REVERSE;
+            if (this.deflect(deflection, null, this.owner, false)) {
+                this.setDeltaMovement(this.getDeltaMovement().scale(0.2));
+                return deflection;
             }
         }
 
@@ -298,11 +_,9 @@
         return false;
     }
 
-    public boolean deflect(
-        ProjectileDeflection deflection, @Nullable Entity deflectingEntity, @Nullable EntityReference<Entity> newOwner, boolean byAttack, Vec3 power
-    ) {
+    public boolean deflect(ProjectileDeflection deflection, @Nullable Entity deflectingEntity, @Nullable EntityReference<Entity> newOwner, boolean byAttack) {
+        deflection.deflect(this, deflectingEntity, this.random);
         if (!this.level().isClientSide()) {
-            deflection.deflect(this, deflectingEntity, this.random, power);
             this.setOwner(newOwner);
             this.onDeflection(byAttack);
         }
@@ -310,16 +_,10 @@
         return true;
     }
 
-    public boolean deflect(
-        ProjectileDeflection deflection, @Nullable Entity deflectingEntity, @Nullable EntityReference<Entity> newOwner, boolean byAttack, double power
-    ) {
-        return this.deflect(deflection, deflectingEntity, newOwner, byAttack, new Vec3(power, power, power));
-    }
-
     protected void onDeflection(boolean byAttack) {
     }
 
-    protected void onItemBreak(ItemStack item) {
+    protected void onItemBreak(Item item) {
     }
 
     protected void onHit(HitResult hitResult) {
@@ -328,7 +_,7 @@
             EntityHitResult entityHitResult = (EntityHitResult)hitResult;
             Entity entityHit = entityHitResult.getEntity();
             if (entityHit.is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entityHit instanceof Projectile projectile) {
-                this.onRedirectProjectile(projectile);
+                projectile.deflect(ProjectileDeflection.AIM_DEFLECT, this.getOwner(), this.owner, true);
             }
 
             this.onHitEntity(entityHitResult);
@@ -345,13 +_,20 @@
     }
 
     protected void onHitBlock(BlockHitResult hitResult) {
+        // CraftBukkit start - cancellable hit event
+        if (hitCancelled) {
+            return;
+        }
+        // CraftBukkit end
         BlockState state = this.level().getBlockState(hitResult.getBlockPos());
         state.onProjectileHit(this.level(), state, hitResult, this);
     }
 
-    protected void onRedirectProjectile(Projectile hitProjectile) {
-        hitProjectile.deflect(ProjectileDeflection.AIM_DEFLECT, this.getOwner(), this.owner, true, 1.0);
+    // Paper start
+    public boolean canHitEntityPublic(final Entity target) {
+        return this.canHitEntity(target);
     }
+    // Paper end
 
     protected boolean canHitEntity(Entity entity) {
         if (!entity.canBeHitByProjectile()) {
@@ -359,6 +_,15 @@
         }
 
         Entity owner = this.getOwner();
+        // Paper start - Cancel hit for vanished entities
+        if (owner instanceof net.minecraft.server.level.ServerPlayer) {
+            org.bukkit.entity.Entity collided = entity.getBukkitEntity();
+            org.bukkit.entity.Player shooter = (org.bukkit.entity.Player) owner.getBukkitEntity();
+            if (!shooter.canSee(collided)) {
+                return false;
+            }
+        }
+        // Paper end - Cancel hit for vanished entities
         return owner == null || this.leftOwner || !owner.isPassengerOfSameVehicle(entity);
     }
 
@@ -370,13 +_,7 @@
     }
 
     protected static float lerpRotation(float rotO, float rot) {
-        while (rot - rotO < -180.0F) {
-            rotO -= 360.0F;
-        }
-
-        while (rot - rotO >= 180.0F) {
-            rotO += 360.0F;
-        }
+        rotO += Math.round((rot - rotO) / 360.0F) * 360.0F; // Paper - stop large look changes from crashing the server
 
         return Mth.lerp(0.2F, rotO, rot);
     }
@@ -399,13 +_,13 @@
     @Override
     public boolean mayInteract(ServerLevel level, BlockPos pos) {
         Entity owner = this.getOwner();
-        return owner instanceof Player ? owner.mayInteract(level, pos) : owner == null || level.getGameRules().get(GameRules.MOB_GRIEFING);
+        return owner instanceof Player ? owner.mayInteract(level, pos) : owner == null || net.neoforged.neoforge.event.EventHooks.canEntityGrief(level, owner);
     }
 
     public boolean mayBreak(ServerLevel level, BlockPos pos) {
         return this.is(EntityTypeTags.IMPACT_PROJECTILES)
             && level.getGameRules().get(GameRules.PROJECTILES_CAN_BREAK_BLOCKS)
-            && this.canBreakBlockInAdventureMode(level, new BlockInWorld(level, pos, false));
+            && this.canBreakBlockInAdventureMode(new BlockInWorld(level, pos, false));
     }
 
     @Override
@@ -439,8 +_,8 @@
     }
 
     @Override
-    public MoveSimulationType getMoveSimulationType() {
-        return MoveSimulationType.SERVER_AND_CLIENT;
+    public boolean canSimulateMovement() {
+        return true;
     }
 
     @FunctionalInterface
