--- a/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -30,17 +_,18 @@
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class ShulkerBullet extends Projectile {
     private static final double SPEED = 0.15;
     @Nullable
     private Entity finalTarget;
     @Nullable
-    private Direction currentMoveDirection;
-    private int flightSteps;
-    private double targetDeltaX;
-    private double targetDeltaY;
-    private double targetDeltaZ;
+    public Direction currentMoveDirection;
+    public int flightSteps;
+    public double targetDeltaX;
+    public double targetDeltaY;
+    public double targetDeltaZ;
     @Nullable
     private UUID targetId;
 
@@ -59,6 +_,18 @@
         this.selectNextMoveDirection(p_37333_);
     }
 
+    // CraftBukkit start
+    public Entity getTarget() {
+        return this.finalTarget;
+    }
+
+    public void setTarget(Entity e) {
+        this.finalTarget = e;
+        this.currentMoveDirection = Direction.UP;
+        this.selectNextMoveDirection(Direction.Axis.X);
+    }
+    // CraftBukkit end
+
     @Override
     public SoundSource getSoundSource() {
         return SoundSource.HOSTILE;
@@ -187,7 +_,7 @@
     @Override
     public void checkDespawn() {
         if (this.level().getDifficulty() == Difficulty.PEACEFUL) {
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         }
     }
 
@@ -218,8 +_,8 @@
             }
 
             HitResult hitresult = ProjectileUtil.getHitResultOnMoveVector(this, this::canHitEntity);
-            if (hitresult.getType() != HitResult.Type.MISS) {
-                this.hitTargetOrDeflectSelf(hitresult);
+            if (hitresult.getType() != HitResult.Type.MISS && !net.neoforged.neoforge.event.EventHooks.onProjectileImpact(this, hitresult)) {
+                this.preHitTargetOrDeflectSelf(hitresult); // CraftBukkit - projectile hit event
             }
         }
 
@@ -255,7 +_,7 @@
     }
 
     @Override
-    protected boolean canHitEntity(Entity p_37341_) {
+    public boolean canHitEntity(Entity p_37341_) {
         return super.canHitEntity(p_37341_) && !p_37341_.noPhysics;
     }
 
@@ -288,6 +_,7 @@
             }
 
             if (entity instanceof LivingEntity livingentity1) {
+                livingentity1.addEffectCause(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK);  // CraftBukkit
                 livingentity1.addEffect(new MobEffectInstance(MobEffects.LEVITATION, 200), MoreObjects.firstNonNull(entity1, this));
             }
         }
@@ -305,10 +_,16 @@
         this.level().gameEvent(GameEvent.ENTITY_DAMAGE, this.position(), GameEvent.Context.of(this));
     }
 
+    private void destroy(org.bukkit.event.entity.EntityRemoveEvent.Cause cause) {
+        this.discard(cause);
+        // CraftBukkit end
+        this.level().gameEvent(GameEvent.ENTITY_DAMAGE, this.position(), GameEvent.Context.of(this));
+    }
+
     @Override
     protected void onHit(HitResult p_37347_) {
         super.onHit(p_37347_);
-        this.destroy();
+        this.destroy(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
     }
 
     @Override
@@ -318,10 +_,15 @@
 
     @Override
     public boolean hurt(DamageSource p_37338_, float p_37339_) {
+        // CraftBukkit start
+        if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_37338_, p_37339_, false)) {
+            return false;
+        }
+        // CraftBukkit end
         if (!this.level().isClientSide) {
             this.playSound(SoundEvents.SHULKER_BULLET_HURT, 1.0F, 1.0F);
             ((ServerLevel)this.level()).sendParticles(ParticleTypes.CRIT, this.getX(), this.getY(), this.getZ(), 15, 0.2, 0.2, 0.2, 0.0);
-            this.destroy();
+            this.destroy(org.bukkit.event.entity.EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause
         }
 
         return true;
