--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -38,6 +_,7 @@
     protected void doPostHurtEffects(LivingEntity p_37422_) {
         super.doPostHurtEffects(p_37422_);
         MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
+        p_37422_.addEffectCause(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
         p_37422_.addEffect(mobeffectinstance, this.getEffectSource());
     }
 
