--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -12,7 +_,7 @@
 import net.minecraft.world.level.Level;
 
 public class SpectralArrow extends AbstractArrow {
-    public int duration = 200;
+    private int duration = 200;
 
     public SpectralArrow(EntityType<? extends SpectralArrow> p_37411_, Level p_37412_) {
         super(p_37411_, p_37412_);
