--- a/net/minecraft/world/entity/monster/breeze/Breeze.java
+++ b/net/minecraft/world/entity/monster/breeze/Breeze.java
@@ -78,7 +78,7 @@
 
     @Override
     public BehaviorController<Breeze> getBrain() {
-        return super.getBrain();
+        return (BehaviorController<Breeze>) super.getBrain(); // CraftBukkit - decompile error
     }
 
     @Override
@@ -246,6 +246,7 @@
 
     @Override
     public boolean canAttackType(EntityTypes<?> entitytypes) {
+        if (this.getTarget() != null) return this.getTarget().getType() == entitytypes; // SPIGOT-7957: Allow attack if target from brain was set
         return entitytypes == EntityTypes.PLAYER || entitytypes == EntityTypes.IRON_GOLEM;
     }
 
