--- a/net/minecraft/world/entity/monster/breeze/Breeze.java
+++ b/net/minecraft/world/entity/monster/breeze/Breeze.java
@@ -78,7 +78,7 @@
 
     @Override
     public Brain<Breeze> getBrain() {
-        return super.getBrain();
+        return (Brain<Breeze>) super.getBrain(); // CraftBukkit - decompile error
     }
 
     @Override
@@ -246,6 +246,7 @@
 
     @Override
     public boolean canAttackType(EntityType<?> entitytype) {
+        if (this.getTarget() != null) return this.getTarget().getType() == entitytype; // SPIGOT-7957: Allow attack if target from brain was set
         return entitytype == EntityType.PLAYER || entitytype == EntityType.IRON_GOLEM;
     }
 
