--- a/net/minecraft/server/notifications/EmptyNotificationService.java
+++ b/net/minecraft/server/notifications/EmptyNotificationService.java
@@ -69,7 +_,7 @@
     }
 
     @Override
-    public <T> void onGameRuleChanged(GameRule<T> gameRule, T value) {
+    public <T> void onGameRuleChanged(net.minecraft.server.level.ServerLevel level, GameRule<T> gameRule, T value) { // Paper - per-world game rules (add level param)
     }
 
     @Override
