--- a/net/minecraft/server/notifications/NotificationManager.java
+++ b/net/minecraft/server/notifications/NotificationManager.java
@@ -107,8 +_,8 @@
     }
 
     @Override
-    public <T> void onGameRuleChanged(GameRule<T> gameRule, T value) {
-        this.notificationServices.forEach(notificationService -> notificationService.onGameRuleChanged(gameRule, value));
+    public <T> void onGameRuleChanged(net.minecraft.server.level.ServerLevel level, GameRule<T> gameRule, T value) { // Paper - per-world game rules (add level param)
+        this.notificationServices.forEach(notificationService -> notificationService.onGameRuleChanged(level, gameRule, value)); // Paper - per-world game rules (add level param)
     }
 
     @Override
