From 27a5d79e63e739b9bf473a1a48fbfdb4af53763d Mon Sep 17 00:00:00 2001
From: erocs <github@erocs.org>
Date: Sun, 8 Sep 2013 12:06:15 -0700
Subject: [PATCH] Hopper Customisations

Allows editing hopper cooldowns and amount transferred per tick.

diff --git a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 68a7c7e5c..0e2a7ffef 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -147,9 +147,14 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         entity.tickedGameTime = level.getGameTime();
         if (!entity.isOnCooldown()) {
             entity.setCooldown(0);
-            tryMoveItems(level, pos, state, entity, () -> {
+            // Spigot start
+            boolean result = tryMoveItems(level, pos, state, entity, () -> {
                 return suckInItems(level, entity);
             });
+            if (!result && entity.level.spigotConfig.hopperCheck > 1) {
+                entity.setCooldown(entity.level.spigotConfig.hopperCheck);
+            }
+            // Spigot end
         }
 
     }
@@ -170,7 +175,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 }
 
                 if (flag) {
-                    entity.setCooldown(8);
+                    entity.setCooldown(level.spigotConfig.hopperTransfer); // Spigot
                     setChanged(level, pos, state);
                     return true;
                 }
@@ -208,7 +213,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                         int j = itemstack.getCount();
                         // CraftBukkit start - Call event when pushing items into other inventories
                         ItemStack original = itemstack.copy();
-                        CraftItemStack oitemstack = CraftItemStack.asCraftMirror(self.removeItem(i, 1));
+                        CraftItemStack oitemstack = CraftItemStack.asCraftMirror(self.removeItem(i, level.spigotConfig.hopperAmount)); // Spigot
 
                         org.bukkit.inventory.Inventory destinationInventory;
                         // Have to special case large chests as they work oddly
@@ -224,9 +229,10 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                         level.getCraftServer().getPluginManager().callEvent(event);
                         if (event.isCancelled()) {
                             self.setItem(i, original);
-                            self.setCooldown(8); // Delay hopper checks
+                            self.setCooldown(level.spigotConfig.hopperTransfer); // Delay hopper checks // Spigot
                             return false;
                         }
+                        int origCount = event.getItem().getAmount(); // Spigot
                         ItemStack itemstack1 = addItem(self, container, CraftItemStack.asNMSCopy(event.getItem()), direction);
                         // CraftBukkit end
 
@@ -236,7 +242,10 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                         }
 
                         itemstack.setCount(j);
-                        if (j == 1) {
+                        // Spigot start
+                        itemstack.shrink(origCount - itemstack1.getCount());
+                        if (j <= level.spigotConfig.hopperAmount) {
+                            // Spigot end
                             self.setItem(i, itemstack);
                         }
                     }
@@ -303,7 +312,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             Direction direction = Direction.DOWN;
 
             for (int i : getSlots(container, direction)) {
-                if (tryTakeInItemFromSlot(hopper, container, i, direction)) {
+                if (tryTakeInItemFromSlot(hopper, container, i, direction, level)) { // Spigot
                     return true;
                 }
             }
@@ -324,14 +333,14 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         }
     }
 
-    private static boolean tryTakeInItemFromSlot(Hopper hopper, Container container, int slot, Direction direction) {
+    private static boolean tryTakeInItemFromSlot(Hopper hopper, Container container, int slot, Direction direction, Level level) { // Spigot
         ItemStack itemstack = container.getItem(slot);
 
         if (!itemstack.isEmpty() && canTakeItemFromContainer(hopper, container, itemstack, slot, direction)) {
             int j = itemstack.getCount();
             // CraftBukkit start - Call event on collection of items from inventories into the hopper
             ItemStack original = itemstack.copy();
-            CraftItemStack oitemstack = CraftItemStack.asCraftMirror(container.removeItem(slot, 1));
+            CraftItemStack oitemstack = CraftItemStack.asCraftMirror(container.removeItem(slot, level.spigotConfig.hopperAmount)); // Spigot
 
             org.bukkit.inventory.Inventory sourceInventory;
             // Have to special case large chests as they work oddly
@@ -350,11 +359,12 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                 container.setItem(slot, original);
 
                 if (hopper instanceof HopperBlockEntity) {
-                    ((HopperBlockEntity) hopper).setCooldown(8); // Delay hopper checks
+                    ((HopperBlockEntity) hopper).setCooldown(level.spigotConfig.hopperTransfer); // Spigot
                 }
 
                 return false;
             }
+            int origCount = event.getItem().getAmount(); // Spigot
             ItemStack itemstack1 = addItem(container, hopper, CraftItemStack.asNMSCopy(event.getItem()), null);
             // CraftBukkit end
 
@@ -364,7 +374,10 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             }
 
             itemstack.setCount(j);
-            if (j == 1) {
+            // Spigot start
+            itemstack.shrink(origCount - itemstack1.getCount());
+            if (j <= level.spigotConfig.hopperAmount) {
+                // Spigot end
                 container.setItem(slot, itemstack);
             }
         }
@@ -465,6 +478,11 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
             boolean flag1 = container.isEmpty();
 
             if (itemstack1.isEmpty()) {
+                // Spigot start - SPIGOT-6693, InventorySubcontainer#setItem
+                if (!itemStack.isEmpty() && itemStack.getCount() > container.getMaxStackSize()) {
+                    itemStack = itemStack.split(container.getMaxStackSize());
+                }
+                // Spigot end
                 container.setItem(slot, itemStack);
                 itemStack = ItemStack.EMPTY;
                 flag = true;
@@ -492,7 +510,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                             }
                         }
 
-                        hopperblockentity.setCooldown(8 - l);
+                        hopperblockentity.setCooldown(hopperblockentity.level.spigotConfig.hopperTransfer - l); // Spigot
                     }
                 }
 
@@ -555,6 +573,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     }
 
     private static @Nullable Container getBlockContainer(Level level, BlockPos pos, BlockState state) {
+        if ( !level.spigotConfig.hopperCanLoadChunks && !level.hasChunkAt( pos ) ) return null; // Spigot
         Block block = state.getBlock();
 
         if (block instanceof WorldlyContainerHolder) {
diff --git a/src/main/java/org/spigotmc/SpigotWorldConfig.java b/src/main/java/org/spigotmc/SpigotWorldConfig.java
index ab984b210..cec948a05 100644
--- a/src/main/java/org/spigotmc/SpigotWorldConfig.java
+++ b/src/main/java/org/spigotmc/SpigotWorldConfig.java
@@ -228,4 +228,22 @@ public class SpigotWorldConfig
         otherTrackingRange = getInt( "entity-tracking-range.other", otherTrackingRange );
         log( "Entity Tracking Range: Pl " + playerTrackingRange + " / An " + animalTrackingRange + " / Mo " + monsterTrackingRange + " / Mi " + miscTrackingRange + " / Di " + displayTrackingRange + " / Other " + otherTrackingRange );
     }
+
+    public int hopperTransfer;
+    public int hopperCheck;
+    public int hopperAmount;
+    public boolean hopperCanLoadChunks;
+    private void hoppers()
+    {
+        // Set the tick delay between hopper item movements
+        hopperTransfer = getInt( "ticks-per.hopper-transfer", 8 );
+        if ( SpigotConfig.version < 11 )
+        {
+            set( "ticks-per.hopper-check", 1 );
+        }
+        hopperCheck = getInt( "ticks-per.hopper-check", 1 );
+        hopperAmount = getInt( "hopper-amount", 1 );
+        hopperCanLoadChunks = getBoolean( "hopper-can-load-chunks", false );
+        log( "Hopper Transfer: " + hopperTransfer + " Hopper Check: " + hopperCheck + " Hopper Amount: " + hopperAmount + " Hopper Can Load Chunks: " + hopperCanLoadChunks );
+    }
 }
-- 
2.53.0

