From 192f67947e188904302e258c19bbc56134ea6da3 Mon Sep 17 00:00:00 2001
From: David <dmck2b@gmail.com>
Date: Mon, 21 Apr 2014 12:43:08 +0100
Subject: [PATCH] Prevent NoClassDefError crash and notify on crash


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index 174c504b7..e0bde3d4e 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -164,6 +164,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     public final org.spigotmc.SpigotWorldConfig spigotConfig; // Spigot
 
     public final SpigotTimings.WorldTimingsHandler timings; // Spigot
+    public static BlockPos lastPhysicsProblem; // Spigot
 
     public CraftWorld getWorld() {
         return this.world;
@@ -339,7 +340,13 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
                 // CraftBukkit start
                 if (!this.captureBlockStates) { // Don't notify clients or update physics while capturing blockstates
                     // Modularize client and physic updates
-                    notifyAndUpdatePhysics(pos, levelchunk, blockstate1, blockState, blockstate2, updateFlags, updateLimit);
+                    // Spigot start
+                    try {
+                        notifyAndUpdatePhysics(pos, levelchunk, blockstate1, blockState, blockstate2, updateFlags, updateLimit);
+                    } catch (StackOverflowError ex) {
+                        lastPhysicsProblem = new BlockPos(pos);
+                    }
+                    // Spigot end
                 }
                 // CraftBukkit end
 
diff --git a/src/main/java/net/minecraft/world/level/redstone/NeighborUpdater.java b/src/main/java/net/minecraft/world/level/redstone/NeighborUpdater.java
index 2d38b6084..fc5b423f7 100644
--- a/src/main/java/net/minecraft/world/level/redstone/NeighborUpdater.java
+++ b/src/main/java/net/minecraft/world/level/redstone/NeighborUpdater.java
@@ -81,6 +81,10 @@ public interface NeighborUpdater {
             }
             // CraftBukkit end
             state.handleNeighborChanged(level, pos, changedBlock, orientation, movedByPiston);
+            // Spigot Start
+        } catch (StackOverflowError ex) {
+            level.lastPhysicsProblem = new BlockPos(pos);
+            // Spigot End
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Exception while updating neighbours");
             CrashReportCategory crashreportcategory = crashreport.addCategory("Block being updated");
diff --git a/src/main/java/org/spigotmc/WatchdogThread.java b/src/main/java/org/spigotmc/WatchdogThread.java
index 4734d4ae8..9a695b80e 100644
--- a/src/main/java/org/spigotmc/WatchdogThread.java
+++ b/src/main/java/org/spigotmc/WatchdogThread.java
@@ -74,6 +74,13 @@ public class WatchdogThread extends Thread
                 log.log( Level.SEVERE, "Be sure to include ALL relevant console errors and Minecraft crash reports" );
                 log.log( Level.SEVERE, "Spigot version: " + Bukkit.getServer().getVersion() );
                 //
+                if ( net.minecraft.world.level.Level.lastPhysicsProblem != null )
+                {
+                    log.log( Level.SEVERE, "------------------------------" );
+                    log.log( Level.SEVERE, "During the run of the server, a physics stackoverflow was supressed" );
+                    log.log( Level.SEVERE, "near " + net.minecraft.world.level.Level.lastPhysicsProblem );
+                }
+                //
                 log.log( Level.SEVERE, "------------------------------" );
                 log.log( Level.SEVERE, "Server thread dump (Look for plugins here before reporting to Spigot!):" );
                 dumpThread( ManagementFactory.getThreadMXBean().getThreadInfo( MinecraftServer.getServer().serverThread.getId(), Integer.MAX_VALUE ), log );
-- 
2.53.0

