From a7860b2c8b56e8ce0fe1a7ec6b1cd4aac48d8c47 Mon Sep 17 00:00:00 2001
From: Suddenly <suddenly@suddenly.coffee>
Date: Tue, 8 Jul 2014 09:44:18 +1000
Subject: [PATCH] Safer JSON Loading


diff --git a/src/main/java/net/minecraft/server/players/CachedUserNameToIdResolver.java b/src/main/java/net/minecraft/server/players/CachedUserNameToIdResolver.java
index 9956189e9..d019201ae 100644
--- a/src/main/java/net/minecraft/server/players/CachedUserNameToIdResolver.java
+++ b/src/main/java/net/minecraft/server/players/CachedUserNameToIdResolver.java
@@ -176,6 +176,11 @@ public class CachedUserNameToIdResolver implements UserNameToIdResolver {
             });
         } catch (FileNotFoundException filenotfoundexception) {
             ;
+        // Spigot Start
+        } catch (com.google.gson.JsonSyntaxException | NullPointerException ex) {
+            CachedUserNameToIdResolver.LOGGER.warn( "Usercache.json is corrupted or has bad formatting. Deleting it to prevent further issues." );
+            this.file.delete();
+        // Spigot End
         } catch (JsonParseException | IOException ioexception) {
             CachedUserNameToIdResolver.LOGGER.warn("Failed to load profile cache {}", this.file, ioexception);
         }
diff --git a/src/main/java/net/minecraft/server/players/StoredUserList.java b/src/main/java/net/minecraft/server/players/StoredUserList.java
index 2acff86fe..33052ee3a 100644
--- a/src/main/java/net/minecraft/server/players/StoredUserList.java
+++ b/src/main/java/net/minecraft/server/players/StoredUserList.java
@@ -171,6 +171,14 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
                         this.map.put(this.getKeyForUser(storeduserentry.getUser()), (V) storeduserentry); // CraftBukkit - decompile error
                     }
                 }
+            // Spigot Start
+            } catch ( com.google.gson.JsonParseException | NullPointerException ex )
+            {
+                org.bukkit.Bukkit.getLogger().log( java.util.logging.Level.WARNING, "Unable to read file " + this.file + ", backing it up to {0}.backup and creating new copy.", ex );
+                File backup = new File( this.file + ".backup" );
+                this.file.renameTo( backup );
+                this.file.delete();
+                // Spigot End
             }
 
         }
-- 
2.53.0

