From 14f2068f903d5b806e5586d63d4ca2f0671347df Mon Sep 17 00:00:00 2001
From: Sander Knauff <sanderknauff@hotmail.com>
Date: Sat, 7 Nov 2020 18:23:42 +1100
Subject: [PATCH] Allow Disabling Player Data Saving


diff --git a/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java b/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
index 93c1b61d8..30be568ce 100644
--- a/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/src/main/java/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -38,6 +38,7 @@ public class PlayerDataStorage {
     }
 
     public void save(Player player) {
+        if (org.spigotmc.SpigotConfig.disablePlayerDataSaving) return; // Spigot
         try (ProblemReporter.ScopedCollector problemreporter_scopedcollector = new ProblemReporter.ScopedCollector(player.problemPath(), PlayerDataStorage.LOGGER)) {
             TagValueOutput tagvalueoutput = TagValueOutput.createWithContext(problemreporter_scopedcollector, player.registryAccess());
 
diff --git a/src/main/java/org/spigotmc/SpigotConfig.java b/src/main/java/org/spigotmc/SpigotConfig.java
index 864967890..898b379e2 100644
--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -396,4 +396,9 @@ public class SpigotConfig
         logVillagerDeaths = getBoolean("settings.log-villager-deaths", true);
         logNamedDeaths = getBoolean("settings.log-named-deaths", true);
     }
+
+    public static boolean disablePlayerDataSaving;
+    private static void disablePlayerDataSaving() {
+        disablePlayerDataSaving = getBoolean("players.disable-saving", false);
+    }
 }
-- 
2.53.0

