--- a/net/minecraft/world/attribute/AttributeTypes.java
+++ b/net/minecraft/world/attribute/AttributeTypes.java
@@ -13,30 +_,27 @@
 import net.minecraft.world.entity.schedule.Activity;
 import net.minecraft.world.level.MoonPhase;
 import net.minecraft.world.level.biome.MobSpawnSettings;
-import org.joml.Vector3fc;
-import org.joml.Vector4fc;
 
 public interface AttributeTypes {
     AttributeType<Boolean> BOOLEAN = register("boolean", AttributeType.ofNotInterpolated(Codec.BOOL, AttributeModifier.BOOLEAN_LIBRARY));
     AttributeType<TriState> TRI_STATE = register("tri_state", AttributeType.ofNotInterpolated(TriState.CODEC));
     AttributeType<Float> FLOAT = register(
-        "float",
-        AttributeType.ofInterpolated(Codec.FLOAT, AttributeModifier.FLOAT_LIBRARY, LerpFunction.ofFloat(), LerpFunction.ofFloat(), value -> value, null)
+        "float", AttributeType.ofInterpolated(Codec.FLOAT, AttributeModifier.FLOAT_LIBRARY, LerpFunction.ofFloat(), LerpFunction.ofFloat(), value -> value)
     );
     AttributeType<Float> ANGLE_DEGREES = register(
         "angle_degrees",
-        AttributeType.ofInterpolated(Codec.FLOAT, AttributeModifier.FLOAT_LIBRARY, LerpFunction.ofFloat(), LerpFunction.ofDegrees(90.0F), value -> value, null)
-    );
-    AttributeType<Vector3fc> RGB_COLOR = register(
-        "rgb_color", AttributeType.ofInterpolated(ExtraCodecs.STRING_RGB_VEC3_COLOR, AttributeModifier.RGB_COLOR_LIBRARY, LerpFunction.ofColorVec3())
-    );
-    AttributeType<Vector4fc> ARGB_COLOR = register(
-        "argb_color", AttributeType.ofInterpolated(ExtraCodecs.STRING_ARGB_VEC4_COLOR, AttributeModifier.ARGB_COLOR_LIBRARY, LerpFunction.ofColorVec4())
+        AttributeType.ofInterpolated(Codec.FLOAT, AttributeModifier.FLOAT_LIBRARY, LerpFunction.ofFloat(), LerpFunction.ofDegrees(90.0F), value -> value)
+    );
+    AttributeType<Integer> RGB_COLOR = register(
+        "rgb_color", AttributeType.ofInterpolated(ExtraCodecs.STRING_RGB_COLOR, AttributeModifier.RGB_COLOR_LIBRARY, LerpFunction.ofColor())
+    );
+    AttributeType<Integer> ARGB_COLOR = register(
+        "argb_color", AttributeType.ofInterpolated(ExtraCodecs.STRING_ARGB_COLOR, AttributeModifier.ARGB_COLOR_LIBRARY, LerpFunction.ofColor())
     );
     AttributeType<Integer> INTEGER = register(
         "integer",
         AttributeType.ofInterpolated(
-            Codec.INT, AttributeModifier.INTEGER_LIBRARY, LerpFunction.ofInteger(), LerpFunction.ofInteger(), value -> value.intValue(), value -> value
+            Codec.INT, AttributeModifier.INTEGER_LIBRARY, LerpFunction.ofInteger(), LerpFunction.ofInteger(), value -> value.intValue()
         )
     );
     AttributeType<MoonPhase> MOON_PHASE = register("moon_phase", AttributeType.ofNotInterpolated(MoonPhase.CODEC));
