--- a/net/minecraft/world/attribute/AttributeType.java
+++ b/net/minecraft/world/attribute/AttributeType.java
@@ -3,7 +_,6 @@
 import com.google.common.collect.ImmutableBiMap;
 import com.mojang.serialization.Codec;
 import java.util.Map;
-import java.util.function.ToIntFunction;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.ToFloatFunction;
@@ -19,13 +_,12 @@
     LerpFunction<Value> stateChangeLerp,
     LerpFunction<Value> spatialLerp,
     LerpFunction<Value> partialTickLerp,
-    @Nullable ToFloatFunction<Value> toFloat,
-    @Nullable ToIntFunction<Value> toInt
+    @Nullable ToFloatFunction<Value> toFloat
 ) {
     public static <Value> AttributeType<Value> ofInterpolated(
         Codec<Value> valueCodec, Map<AttributeModifier.OperationId, AttributeModifier<Value, ?>> modifierLibrary, LerpFunction<Value> lerp
     ) {
-        return ofInterpolated(valueCodec, modifierLibrary, lerp, lerp, null, null);
+        return ofInterpolated(valueCodec, modifierLibrary, lerp, lerp, null);
     }
 
     public static <Value> AttributeType<Value> ofInterpolated(
@@ -33,10 +_,9 @@
         Map<AttributeModifier.OperationId, AttributeModifier<Value, ?>> modifierLibrary,
         LerpFunction<Value> lerp,
         LerpFunction<Value> partialTickLerp,
-        @Nullable ToFloatFunction<Value> toFloat,
-        @Nullable ToIntFunction<Value> toInt
+        @Nullable ToFloatFunction<Value> toFloat
     ) {
-        return new AttributeType<>(valueCodec, modifierLibrary, createModifierCodec(modifierLibrary), lerp, lerp, lerp, partialTickLerp, toFloat, toInt);
+        return new AttributeType<>(valueCodec, modifierLibrary, createModifierCodec(modifierLibrary), lerp, lerp, lerp, partialTickLerp, toFloat);
     }
 
     public static <Value> AttributeType<Value> ofNotInterpolated(
@@ -50,7 +_,6 @@
             LerpFunction.ofStep(0.0F),
             LerpFunction.ofStep(0.5F),
             LerpFunction.ofStep(0.0F),
-            null,
             null
         );
     }
@@ -78,14 +_,6 @@
             throw new IllegalStateException(value + " cannot be represented as a float");
         } else {
             return this.toFloat.applyAsFloat(value);
-        }
-    }
-
-    public int toInt(Value value) {
-        if (this.toInt == null) {
-            throw new IllegalStateException(value + " cannot be represented as an integer");
-        } else {
-            return this.toInt.applyAsInt(value);
         }
     }
 
