--- a/net/minecraft/world/attribute/EnvironmentAttributes.java
+++ b/net/minecraft/world/attribute/EnvironmentAttributes.java
@@ -7,17 +_,14 @@
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.resources.Identifier;
-import net.minecraft.util.ARGB;
 import net.minecraft.util.TriState;
 import net.minecraft.world.entity.schedule.Activity;
 import net.minecraft.world.level.MoonPhase;
 import net.minecraft.world.level.biome.MobSpawnSettings;
-import org.joml.Vector3fc;
-import org.joml.Vector4fc;
 
 public interface EnvironmentAttributes {
-    EnvironmentAttribute<Vector3fc> FOG_COLOR = register(
-        "visual/fog_color", EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(ARGB.vector3fFromRGB24(0)).spatiallyInterpolated().syncable()
+    EnvironmentAttribute<Integer> FOG_COLOR = register(
+        "visual/fog_color", EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(0).spatiallyInterpolated().syncable()
     );
     EnvironmentAttribute<Float> FOG_START_DISTANCE = register(
         "visual/fog_start_distance", EnvironmentAttribute.builder(AttributeTypes.FLOAT).defaultValue(0.0F).spatiallyInterpolated().syncable()
@@ -46,9 +_,8 @@
             .spatiallyInterpolated()
             .syncable()
     );
-    EnvironmentAttribute<Vector3fc> WATER_FOG_COLOR = register(
-        "visual/water_fog_color",
-        EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(ARGB.vector3fFromRGB24(-16448205)).spatiallyInterpolated().syncable()
+    EnvironmentAttribute<Integer> WATER_FOG_COLOR = register(
+        "visual/water_fog_color", EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(-16448205).spatiallyInterpolated().syncable()
     );
     EnvironmentAttribute<Float> WATER_FOG_START_DISTANCE = register(
         "visual/water_fog_start_distance", EnvironmentAttribute.builder(AttributeTypes.FLOAT).defaultValue(-8.0F).spatiallyInterpolated().syncable()
@@ -57,16 +_,14 @@
         "visual/water_fog_end_distance",
         EnvironmentAttribute.builder(AttributeTypes.FLOAT).defaultValue(96.0F).valueRange(AttributeRange.NON_NEGATIVE_FLOAT).spatiallyInterpolated().syncable()
     );
-    EnvironmentAttribute<Vector3fc> SKY_COLOR = register(
-        "visual/sky_color", EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(ARGB.vector3fFromRGB24(0)).spatiallyInterpolated().syncable()
-    );
-    EnvironmentAttribute<Vector4fc> SUNRISE_SUNSET_COLOR = register(
-        "visual/sunrise_sunset_color",
-        EnvironmentAttribute.builder(AttributeTypes.ARGB_COLOR).defaultValue(ARGB.vector4fFromARGB32(0)).spatiallyInterpolated().syncable()
-    );
-    EnvironmentAttribute<Vector4fc> CLOUD_COLOR = register(
-        "visual/cloud_color",
-        EnvironmentAttribute.builder(AttributeTypes.ARGB_COLOR).defaultValue(ARGB.vector4fFromARGB32(0)).spatiallyInterpolated().syncable()
+    EnvironmentAttribute<Integer> SKY_COLOR = register(
+        "visual/sky_color", EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(0).spatiallyInterpolated().syncable()
+    );
+    EnvironmentAttribute<Integer> SUNRISE_SUNSET_COLOR = register(
+        "visual/sunrise_sunset_color", EnvironmentAttribute.builder(AttributeTypes.ARGB_COLOR).defaultValue(0).spatiallyInterpolated().syncable()
+    );
+    EnvironmentAttribute<Integer> CLOUD_COLOR = register(
+        "visual/cloud_color", EnvironmentAttribute.builder(AttributeTypes.ARGB_COLOR).defaultValue(0).spatiallyInterpolated().syncable()
     );
     EnvironmentAttribute<Float> CLOUD_HEIGHT = register(
         "visual/cloud_height", EnvironmentAttribute.builder(AttributeTypes.FLOAT).defaultValue(192.33F).spatiallyInterpolated().syncable()
@@ -87,25 +_,21 @@
         "visual/star_brightness",
         EnvironmentAttribute.builder(AttributeTypes.FLOAT).defaultValue(0.0F).valueRange(AttributeRange.UNIT_FLOAT).spatiallyInterpolated().syncable()
     );
-    EnvironmentAttribute<Vector3fc> BLOCK_LIGHT_TINT = register(
-        "visual/block_light_tint",
-        EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(ARGB.vector3fFromRGB24(-10100)).spatiallyInterpolated().syncable()
+    EnvironmentAttribute<Integer> BLOCK_LIGHT_TINT = register(
+        "visual/block_light_tint", EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(-10100).spatiallyInterpolated().syncable()
     );
-    EnvironmentAttribute<Vector3fc> SKY_LIGHT_COLOR = register(
-        "visual/sky_light_color",
-        EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(ARGB.vector3fFromRGB24(-1)).spatiallyInterpolated().syncable()
+    EnvironmentAttribute<Integer> SKY_LIGHT_COLOR = register(
+        "visual/sky_light_color", EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(-1).spatiallyInterpolated().syncable()
     );
     EnvironmentAttribute<Float> SKY_LIGHT_FACTOR = register(
         "visual/sky_light_factor",
         EnvironmentAttribute.builder(AttributeTypes.FLOAT).defaultValue(1.0F).valueRange(AttributeRange.UNIT_FLOAT).spatiallyInterpolated().syncable()
     );
-    EnvironmentAttribute<Vector3fc> NIGHT_VISION_COLOR = register(
-        "visual/night_vision_color",
-        EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(ARGB.vector3fFromRGB24(-6710887)).spatiallyInterpolated().syncable()
+    EnvironmentAttribute<Integer> NIGHT_VISION_COLOR = register(
+        "visual/night_vision_color", EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(-6710887).spatiallyInterpolated().syncable()
     );
-    EnvironmentAttribute<Vector3fc> AMBIENT_LIGHT_COLOR = register(
-        "visual/ambient_light_color",
-        EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(ARGB.vector3fFromRGB24(-16777216)).spatiallyInterpolated().syncable()
+    EnvironmentAttribute<Integer> AMBIENT_LIGHT_COLOR = register(
+        "visual/ambient_light_color", EnvironmentAttribute.builder(AttributeTypes.RGB_COLOR).defaultValue(-16777216).spatiallyInterpolated().syncable()
     );
     EnvironmentAttribute<ParticleOptions> DEFAULT_DRIPSTONE_PARTICLE = register(
         "visual/default_dripstone_particle",
@@ -181,8 +_,7 @@
         "gameplay/can_pillager_patrol_spawn", EnvironmentAttribute.builder(AttributeTypes.BOOLEAN).defaultValue(true)
     );
     EnvironmentAttribute<MobSpawnSettings> NATURAL_MOB_SPAWNS = register(
-        "gameplay/natural_mob_spawns",
-        EnvironmentAttribute.builder(AttributeTypes.MOB_SPAWN_SETTINGS).defaultValue(MobSpawnSettings.EMPTY).fullResolutionBiomes()
+        "gameplay/natural_mob_spawns", EnvironmentAttribute.builder(AttributeTypes.MOB_SPAWN_SETTINGS).defaultValue(MobSpawnSettings.EMPTY)
     );
     EnvironmentAttribute<Float> CREATURE_WORLD_GEN_SPAWN_PROBABILITY = register(
         "gameplay/creature_world_gen_spawn_probability",
