--- a/net/minecraft/world/attribute/EnvironmentAttributeMap.java
+++ b/net/minecraft/world/attribute/EnvironmentAttributeMap.java
@@ -78,9 +_,9 @@
     }
 
     public static class Builder {
-        protected final Map<EnvironmentAttribute<?>, EnvironmentAttributeMap.Entry<?, ?>> entries = new HashMap<>();
+        private final Map<EnvironmentAttribute<?>, EnvironmentAttributeMap.Entry<?, ?>> entries = new HashMap<>();
 
-        protected Builder() {
+        private Builder() {
         }
 
         public EnvironmentAttributeMap.Builder putAll(EnvironmentAttributeMap map) {
