--- a/net/minecraft/world/attribute/EnvironmentAttribute.java
+++ b/net/minecraft/world/attribute/EnvironmentAttribute.java
@@ -13,7 +_,6 @@
     private final boolean isSyncable;
     private final boolean isPositional;
     private final boolean isSpatiallyInterpolated;
-    private final boolean fullResolutionBiomes;
 
     private EnvironmentAttribute(
         AttributeType<Value> type,
@@ -21,8 +_,7 @@
         AttributeRange<Value> valueRange,
         boolean isSyncable,
         boolean isPositional,
-        boolean isSpatiallyInterpolated,
-        boolean fullResolutionBiomes
+        boolean isSpatiallyInterpolated
     ) {
         this.type = type;
         this.defaultValue = defaultValue;
@@ -30,7 +_,6 @@
         this.isSyncable = isSyncable;
         this.isPositional = isPositional;
         this.isSpatiallyInterpolated = isSpatiallyInterpolated;
-        this.fullResolutionBiomes = fullResolutionBiomes;
     }
 
     public static <Value> EnvironmentAttribute.Builder<Value> builder(AttributeType<Value> type) {
@@ -65,10 +_,6 @@
         return this.isSpatiallyInterpolated;
     }
 
-    public boolean isFullResolutionBiomes() {
-        return this.fullResolutionBiomes;
-    }
-
     @Override
     public String toString() {
         return Util.getRegisteredName(BuiltInRegistries.ENVIRONMENT_ATTRIBUTE, this);
@@ -81,7 +_,6 @@
         private boolean isSyncable = false;
         private boolean isPositional = true;
         private boolean isSpatiallyInterpolated = false;
-        private boolean fullResolutionBiomes = false;
 
         public Builder(AttributeType<Value> type) {
             this.type = type;
@@ -112,12 +_,6 @@
             return this;
         }
 
-        @Deprecated
-        public EnvironmentAttribute.Builder<Value> fullResolutionBiomes() {
-            this.fullResolutionBiomes = true;
-            return this;
-        }
-
         public EnvironmentAttribute<Value> build() {
             return new EnvironmentAttribute<>(
                 this.type,
@@ -125,8 +_,7 @@
                 this.valueRange,
                 this.isSyncable,
                 this.isPositional,
-                this.isSpatiallyInterpolated,
-                this.fullResolutionBiomes
+                this.isSpatiallyInterpolated
             );
         }
     }
