--- a/net/minecraft/world/attribute/LerpFunction.java
+++ b/net/minecraft/world/attribute/LerpFunction.java
@@ -5,8 +_,6 @@
 import java.util.List;
 import net.minecraft.util.ARGB;
 import net.minecraft.util.Mth;
-import org.joml.Vector3fc;
-import org.joml.Vector4fc;
 
 public interface LerpFunction<T> {
     LerpFunction<?> CONSTANT = ofStep(1.0F);
@@ -35,14 +_,6 @@
     }
 
     static LerpFunction<Integer> ofColor() {
-        return ARGB::srgbLerp;
-    }
-
-    static LerpFunction<Vector3fc> ofColorVec3() {
-        return ARGB::srgbLerp;
-    }
-
-    static LerpFunction<Vector4fc> ofColorVec4() {
         return ARGB::srgbLerp;
     }
 
