--- a/net/minecraft/world/attribute/WeatherAttributes.java
+++ b/net/minecraft/world/attribute/WeatherAttributes.java
@@ -12,7 +_,7 @@
 public class WeatherAttributes {
     public static final EnvironmentAttributeMap RAIN = EnvironmentAttributeMap.builder()
         .modify(EnvironmentAttributes.SKY_COLOR, ColorModifier.BLEND_TO_GRAY_RGB, new ColorModifier.BlendToGray(0.6F, 0.75F))
-        .modify(EnvironmentAttributes.FOG_COLOR, ColorModifier.MULTIPLY_RGB, ARGB.vector3fFromRGB24(ARGB.colorFromFloat(1.0F, 0.5F, 0.5F, 0.6F)))
+        .modify(EnvironmentAttributes.FOG_COLOR, ColorModifier.MULTIPLY_RGB, ARGB.vector3fFromRGB24(ARGB.colorFromFloat(1.0F, 0.5F, 0.5F, 0.7F)))
         .modify(EnvironmentAttributes.CLOUD_COLOR, ColorModifier.BLEND_TO_GRAY_ARGB, new ColorModifier.BlendToGray(0.24F, 0.5F))
         .modify(EnvironmentAttributes.SKY_LIGHT_LEVEL, FloatModifier.ALPHA_BLEND, new FloatWithAlpha(4.0F, 0.3125F))
         .modify(
@@ -25,7 +_,7 @@
         .build();
     public static final EnvironmentAttributeMap THUNDER = EnvironmentAttributeMap.builder()
         .modify(EnvironmentAttributes.SKY_COLOR, ColorModifier.BLEND_TO_GRAY_RGB, new ColorModifier.BlendToGray(0.24F, 0.94F))
-        .modify(EnvironmentAttributes.FOG_COLOR, ColorModifier.MULTIPLY_RGB, ARGB.vector3fFromRGB24(ARGB.colorFromFloat(1.0F, 0.25F, 0.25F, 0.3F)))
+        .modify(EnvironmentAttributes.FOG_COLOR, ColorModifier.MULTIPLY_RGB, ARGB.vector3fFromRGB24(ARGB.colorFromFloat(1.0F, 0.25F, 0.25F, 0.55F)))
         .modify(EnvironmentAttributes.CLOUD_COLOR, ColorModifier.BLEND_TO_GRAY_ARGB, new ColorModifier.BlendToGray(0.095F, 0.94F))
         .modify(EnvironmentAttributes.SKY_LIGHT_LEVEL, FloatModifier.ALPHA_BLEND, new FloatWithAlpha(4.0F, 0.52734375F))
         .modify(
