--- a/net/minecraft/world/waypoints/WaypointTransmitter.java
+++ b/net/minecraft/world/waypoints/WaypointTransmitter.java
@@ -21,6 +21,11 @@
     Waypoint.a waypointIcon();
 
     static boolean doesSourceIgnoreReceiver(EntityLiving entityliving, EntityPlayer entityplayer) {
+        // CraftBukkit start
+        if (!entityplayer.getBukkitEntity().canSee(entityliving.getBukkitEntity())) {
+            return true;
+        }
+        // CraftBukkit end
         if (entityplayer.isSpectator()) {
             return false;
         } else if (!entityliving.isSpectator() && !entityliving.hasIndirectPassenger(entityplayer)) {
