--- a/net/minecraft/world/waypoints/WaypointTransmitter.java
+++ b/net/minecraft/world/waypoints/WaypointTransmitter.java
@@ -21,6 +21,11 @@
     Waypoint.Icon waypointIcon();
 
     static boolean doesSourceIgnoreReceiver(LivingEntity livingentity, ServerPlayer serverplayer) {
+        // CraftBukkit start
+        if (!serverplayer.getBukkitEntity().canSee(livingentity.getBukkitEntity())) {
+            return true;
+        }
+        // CraftBukkit end
         if (serverplayer.isSpectator()) {
             return false;
         } else if (!livingentity.isSpectator() && !livingentity.hasIndirectPassenger(serverplayer)) {
