--- a/net/minecraft/world/level/gamerules/GameRules.java
+++ b/net/minecraft/world/level/gamerules/GameRules.java
@@ -14,6 +_,7 @@
 import net.minecraft.core.Registry;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.flag.FeatureFlagSet;
 import net.minecraft.world.flag.FeatureFlags;
 import org.jspecify.annotations.Nullable;
@@ -121,6 +_,13 @@
         return this.rules.keySet().stream();
     }
 
+    public <T> boolean get(GameRule<T> rule, Boolean gameRuleOverride) {
+        if (gameRuleOverride != null) {
+            return gameRuleOverride;
+        }
+        return (Boolean) this.get(rule);
+    }
+
     public <T> T get(GameRule<T> gameRule) {
         T value = this.rules.get(gameRule);
         if (value == null) {
@@ -130,31 +_,43 @@
         }
     }
 
-    public <T> void set(GameRule<T> gameRule, T value, @Nullable MinecraftServer server) {
+    public <T> void set(GameRule<T> gameRule, T value, @Nullable ServerLevel level) { // Paper - per-world game rules
         if (!this.rules.has(gameRule)) {
             LOGGER.warn("Tried to set invalid game rule '{}' to value '{}'", gameRule.getIdentifierWithFallback(), value);
         } else {
             this.rules.set(gameRule, value);
-            if (server != null) {
-                server.onGameRuleChanged(gameRule, value);
+            // Paper start - per-world game rules
+            if (level != null) {
+                level.getServer().onGameRuleChanged(level, gameRule, value);
+                // Paper end - per-world game rules
             }
         }
+    }
+
+    public <T> void set(GameRule<T> gameRule, T value, @Nullable MinecraftServer server) {
+      set(gameRule, value, server == null ? null : server.overworld());
     }
 
     public GameRules copy(FeatureFlagSet enabledFeatures) {
         return new GameRules(enabledFeatures, GameRuleMap.copyOf(this.rules));
     }
 
-    public void setAll(GameRules other, @Nullable MinecraftServer server) {
-        this.setAll(other.rules, server);
-    }
-
-    public void setAll(GameRuleMap gameRulesMap, @Nullable MinecraftServer server) {
-        gameRulesMap.keySet().forEach(gameRule -> this.setFromOther(gameRulesMap, (GameRule<?>)gameRule, server));
-    }
-
-    private <T> void setFromOther(GameRuleMap gameRulesMap, GameRule<T> gameRule, @Nullable MinecraftServer server) {
-        this.set(gameRule, Objects.requireNonNull(gameRulesMap.get(gameRule)), server);
+    // Paper start - per-world game rules
+    public void setAll(final GameRules other, final net.minecraft.server.level.@Nullable ServerLevel level) {
+        this.setAll(other.rules, level);
+        // Paper end - per-world game rules
+    }
+
+    // Paper start - per-world game rules
+    public void setAll(final GameRuleMap gameRulesMap, final net.minecraft.server.level.@Nullable ServerLevel level) {
+        gameRulesMap.keySet().forEach(gameRule -> this.setFromOther(gameRulesMap, (GameRule<?>)gameRule, level));
+        // Paper end - per-world game rules
+    }
+
+    // Paper start - per-world game rules
+    private <T> void setFromOther(final GameRuleMap gameRulesMap, final GameRule<T> gameRule, final net.minecraft.server.level.@Nullable ServerLevel level) {
+        this.set(gameRule, Objects.requireNonNull(gameRulesMap.get(gameRule)), level);
+        // Paper end - per-world game rules
     }
 
     public void visitGameRuleTypes(GameRuleTypeVisitor visitor) {
