--- a/net/minecraft/commands/arguments/blocks/BlockInput.java
+++ b/net/minecraft/commands/arguments/blocks/BlockInput.java
@@ -23,7 +_,7 @@
     private static final Logger LOGGER = LogUtils.getLogger();
     private final BlockState state;
     private final Set<Property<?>> properties;
-    public final @Nullable CompoundTag tag;
+    private final @Nullable CompoundTag tag;
 
     public BlockInput(BlockState state, Set<Property<?>> properties, @Nullable CompoundTag tag) {
         this.state = state;
