--- a/net/minecraft/commands/arguments/blocks/BlockInput.java
+++ b/net/minecraft/commands/arguments/blocks/BlockInput.java
@@ -17,7 +_,7 @@
     private final BlockState state;
     private final Set<Property<?>> properties;
     @Nullable
-    private final CompoundTag tag;
+    public final CompoundTag tag;
 
     public BlockInput(BlockState p_114666_, Set<Property<?>> p_114667_, @Nullable CompoundTag p_114668_) {
         this.state = p_114666_;
