--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -188,6 +188,12 @@
         return this.status == Raid.RaidStatus.LOSS;
     }
 
+    // CraftBukkit start
+    public boolean isInProgress() {
+        return this.status == RaidStatus.ONGOING;
+    }
+    // CraftBukkit end
+
     public float getTotalHealth() {
         return this.totalHealth;
     }
@@ -278,6 +284,7 @@
 
                 this.active = level.hasChunkAt(this.center);
                 if (level.getDifficulty() == Difficulty.PEACEFUL) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.PEACE); // CraftBukkit
                     this.stop();
                     return;
                 }
@@ -297,13 +304,16 @@
                 if (!level.isVillage(this.center)) {
                     if (this.groupsSpawned > 0) {
                         this.status = Raid.RaidStatus.LOSS;
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidFinishEvent(this, level, new java.util.ArrayList<>()); // CraftBukkit
                     } else {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.NOT_IN_VILLAGE); // CraftBukkit
                         this.stop();
                     }
                 }
 
                 ++this.ticksActive;
                 if (this.ticksActive >= 48000L) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.TIMEOUT); // CraftBukkit
                     this.stop();
                     return;
                 }
@@ -380,6 +390,7 @@
                     }
 
                     if (j > 5) {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.UNSPAWNABLE);  // CraftBukkit
                         this.stop();
                         break;
                     }
@@ -391,6 +402,7 @@
                     } else {
                         this.status = Raid.RaidStatus.VICTORY;
 
+                        List<org.bukkit.entity.Player> winners = new java.util.ArrayList<>(); // CraftBukkit
                         for (UUID uuid : this.heroesOfTheVillage) {
                             Entity entity = level.getEntity(uuid);
 
@@ -404,10 +416,12 @@
 
                                         serverplayer.awardStat(Stats.RAID_WIN);
                                         CriteriaTriggers.RAID_WIN.trigger(serverplayer);
+                                        winners.add(serverplayer.getBukkitEntity()); // CraftBukkit
                                     }
                                 }
                             }
                         }
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidFinishEvent(this, level, winners); // CraftBukkit
                     }
                 }
 
@@ -415,6 +429,7 @@
             } else if (this.isOver()) {
                 ++this.celebrationTicks;
                 if (this.celebrationTicks >= 600) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.FINISHED); // CraftBukkit
                     this.stop();
                     return;
                 }
@@ -536,6 +551,10 @@
         DifficultyInstance difficultyinstance = level.getCurrentDifficultyAt(pos);
         boolean flag1 = this.shouldSpawnBonusGroup();
 
+        // CraftBukkit start
+        Raider leader = null;
+        List<Raider> raiders = new java.util.ArrayList<>();
+        // CraftBukkit end
         for (Raid.RaiderType raid_raidertype : Raid.RaiderType.VALUES) {
             int j = this.getDefaultNumSpawns(raid_raidertype, i, flag1) + this.getPotentialBonusSpawns(raid_raidertype, this.random, i, difficultyinstance, flag1);
             int k = 0;
@@ -551,9 +570,11 @@
                     raider.setPatrolLeader(true);
                     this.setLeader(i, raider);
                     flag = true;
+                    leader = raider; // CraftBukkit
                 }
 
                 this.joinRaid(level, i, raider, pos, false);
+                raiders.add(raider); // CraftBukkit
                 if (raid_raidertype.entityType == EntityType.RAVAGER) {
                     Raider raider1 = null;
 
@@ -572,6 +593,7 @@
                         this.joinRaid(level, i, raider1, pos, false);
                         raider1.snapTo(pos, 0.0F, 0.0F);
                         raider1.startRiding(raider, false, false);
+                        raiders.add(raider); // CraftBukkit
                     }
                 }
             }
@@ -581,6 +603,7 @@
         ++this.groupsSpawned;
         this.updateBossbar();
         this.setDirty(level);
+        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidSpawnWaveEvent(this, level, leader, raiders); // CraftBukkit
     }
 
     public void joinRaid(ServerLevel level, int groupNumber, Raider raider, @Nullable BlockPos pos, boolean exists) {
@@ -596,7 +619,7 @@
                 raider.finalizeSpawn(level, level.getCurrentDifficultyAt(pos), EntitySpawnReason.EVENT, (SpawnGroupData) null);
                 raider.applyRaidBuffs(level, groupNumber, false);
                 raider.setOnGround(true);
-                level.addFreshEntityWithPassengers(raider);
+                level.addFreshEntityWithPassengers(raider, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.RAID); // CraftBukkit
             }
         }
 
@@ -825,6 +848,12 @@
         this.heroesOfTheVillage.add(killer.getUUID());
     }
 
+    // CraftBukkit start - a method to get all raiders
+    public java.util.Collection<Raider> getRaiders() {
+        return this.groupRaiderMap.values().stream().flatMap(Set::stream).collect(java.util.stream.Collectors.toSet());
+    }
+    // CraftBukkit end
+
     private static enum RaidStatus implements StringRepresentable {
 
         ONGOING("ongoing"), VICTORY("victory"), LOSS("loss"), STOPPED("stopped");
