--- a/net/minecraft/data/advancements/packs/VanillaStoryAdvancements.java
+++ b/net/minecraft/data/advancements/packs/VanillaStoryAdvancements.java
@@ -40,8 +_,8 @@
     @Override
     public void generate() {
         AdvancementHolder root = Advancement.Builder.advancement()
-            .rootDisplay(
-                Items.GRASS_BLOCK,
+            .display(
+                Blocks.GRASS_BLOCK,
                 Component.translatable("advancements.story.root.title"),
                 Component.translatable("advancements.story.root.description"),
                 Identifier.withDefaultNamespace("gui/advancements/backgrounds/stone"),
@@ -58,6 +_,7 @@
                 Items.WOODEN_PICKAXE,
                 Component.translatable("advancements.story.mine_stone.title"),
                 Component.translatable("advancements.story.mine_stone.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -73,6 +_,7 @@
                 Items.STONE_PICKAXE,
                 Component.translatable("advancements.story.upgrade_tools.title"),
                 Component.translatable("advancements.story.upgrade_tools.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -86,6 +_,7 @@
                 Items.IRON_INGOT,
                 Component.translatable("advancements.story.smelt_iron.title"),
                 Component.translatable("advancements.story.smelt_iron.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -99,6 +_,7 @@
                 Items.IRON_PICKAXE,
                 Component.translatable("advancements.story.iron_tools.title"),
                 Component.translatable("advancements.story.iron_tools.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -112,6 +_,7 @@
                 Items.DIAMOND,
                 Component.translatable("advancements.story.mine_diamond.title"),
                 Component.translatable("advancements.story.mine_diamond.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -125,6 +_,7 @@
                 Items.LAVA_BUCKET,
                 Component.translatable("advancements.story.lava_bucket.title"),
                 Component.translatable("advancements.story.lava_bucket.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -138,6 +_,7 @@
                 Items.IRON_CHESTPLATE,
                 Component.translatable("advancements.story.obtain_armor.title"),
                 Component.translatable("advancements.story.obtain_armor.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -155,6 +_,7 @@
                 Items.ENCHANTED_BOOK,
                 Component.translatable("advancements.story.enchant_item.title"),
                 Component.translatable("advancements.story.enchant_item.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -165,9 +_,10 @@
         AdvancementHolder formObsidian = Advancement.Builder.advancement()
             .parent(lavaBucket)
             .display(
-                Items.OBSIDIAN,
+                Blocks.OBSIDIAN,
                 Component.translatable("advancements.story.form_obsidian.title"),
                 Component.translatable("advancements.story.form_obsidian.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -181,6 +_,7 @@
                 Items.SHIELD,
                 Component.translatable("advancements.story.deflect_arrow.title"),
                 Component.translatable("advancements.story.deflect_arrow.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -197,6 +_,7 @@
                 Items.DIAMOND_CHESTPLATE,
                 Component.translatable("advancements.story.shiny_gear.title"),
                 Component.translatable("advancements.story.shiny_gear.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -214,6 +_,7 @@
                 Items.FLINT_AND_STEEL,
                 Component.translatable("advancements.story.enter_the_nether.title"),
                 Component.translatable("advancements.story.enter_the_nether.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -227,6 +_,7 @@
                 Items.GOLDEN_APPLE,
                 Component.translatable("advancements.story.cure_zombie_villager.title"),
                 Component.translatable("advancements.story.cure_zombie_villager.description"),
+                null,
                 AdvancementType.GOAL,
                 true,
                 true,
@@ -240,6 +_,7 @@
                 Items.ENDER_EYE,
                 Component.translatable("advancements.story.follow_ender_eye.title"),
                 Component.translatable("advancements.story.follow_ender_eye.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -253,9 +_,10 @@
         Advancement.Builder.advancement()
             .parent(followEnderEye)
             .display(
-                Items.END_STONE,
+                Blocks.END_STONE,
                 Component.translatable("advancements.story.enter_the_end.title"),
                 Component.translatable("advancements.story.enter_the_end.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
