--- a/net/minecraft/data/advancements/packs/VanillaTheEndAdvancements.java
+++ b/net/minecraft/data/advancements/packs/VanillaTheEndAdvancements.java
@@ -45,8 +_,8 @@
     @Override
     public void generate() {
         AdvancementHolder root = Advancement.Builder.advancement()
-            .rootDisplay(
-                Items.END_STONE,
+            .display(
+                Blocks.END_STONE,
                 Component.translatable("advancements.end.root.title"),
                 Component.translatable("advancements.end.root.description"),
                 Identifier.withDefaultNamespace("gui/advancements/backgrounds/end"),
@@ -60,9 +_,10 @@
         AdvancementHolder killDragon = Advancement.Builder.advancement()
             .parent(root)
             .display(
-                Items.DRAGON_HEAD,
+                Blocks.DRAGON_HEAD,
                 Component.translatable("advancements.end.kill_dragon.title"),
                 Component.translatable("advancements.end.kill_dragon.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -79,6 +_,7 @@
                 Items.ENDER_PEARL,
                 Component.translatable("advancements.end.enter_end_gateway.title"),
                 Component.translatable("advancements.end.enter_end_gateway.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -92,6 +_,7 @@
                 Items.END_CRYSTAL,
                 Component.translatable("advancements.end.respawn_dragon.title"),
                 Component.translatable("advancements.end.respawn_dragon.description"),
+                null,
                 AdvancementType.GOAL,
                 true,
                 true,
@@ -105,9 +_,10 @@
         AdvancementHolder findEndCity = Advancement.Builder.advancement()
             .parent(enterEndGateway)
             .display(
-                Items.PURPUR_BLOCK,
+                Blocks.PURPUR_BLOCK,
                 Component.translatable("advancements.end.find_end_city.title"),
                 Component.translatable("advancements.end.find_end_city.description"),
+                null,
                 AdvancementType.TASK,
                 true,
                 true,
@@ -123,6 +_,7 @@
                 Items.DRAGON_BREATH,
                 Component.translatable("advancements.end.dragon_breath.title"),
                 Component.translatable("advancements.end.dragon_breath.description"),
+                null,
                 AdvancementType.GOAL,
                 true,
                 true,
@@ -136,6 +_,7 @@
                 Items.SHULKER_SHELL,
                 Component.translatable("advancements.end.levitate.title"),
                 Component.translatable("advancements.end.levitate.description"),
+                null,
                 AdvancementType.CHALLENGE,
                 true,
                 true,
@@ -150,6 +_,7 @@
                 Items.ELYTRA,
                 Component.translatable("advancements.end.elytra.title"),
                 Component.translatable("advancements.end.elytra.description"),
+                null,
                 AdvancementType.GOAL,
                 true,
                 true,
@@ -160,9 +_,10 @@
         Advancement.Builder.advancement()
             .parent(killDragon)
             .display(
-                Items.DRAGON_EGG,
+                Blocks.DRAGON_EGG,
                 Component.translatable("advancements.end.dragon_egg.title"),
                 Component.translatable("advancements.end.dragon_egg.description"),
+                null,
                 AdvancementType.GOAL,
                 true,
                 true,
