--- a/net/minecraft/world/item/component/ConsumableListener.java
+++ b/net/minecraft/world/item/component/ConsumableListener.java
@@ -7,4 +7,6 @@
 public interface ConsumableListener {
 
     void onConsume(Level level, LivingEntity livingentity, ItemStack itemstack, Consumable consumable);
+
+    default void cancelUsingItem(net.minecraft.server.level.ServerPlayer serverplayer, ItemStack itemstack) {} // CraftBukkit
 }
