--- a/net/minecraft/client/gui/screens/options/InWorldGameRulesScreen.java
+++ b/net/minecraft/client/gui/screens/options/InWorldGameRulesScreen.java
@@ -15,6 +_,7 @@
 import net.minecraft.network.protocol.game.ServerboundClientCommandPacket;
 import net.minecraft.network.protocol.game.ServerboundSetGameRulePacket;
 import net.minecraft.resources.ResourceKey;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.level.gamerules.GameRule;
 import net.minecraft.world.level.gamerules.GameRuleMap;
 import net.minecraft.world.level.gamerules.GameRules;
@@ -112,7 +_,7 @@
     private <T> void initializeGameRuleValue(GameRule<T> rule, String valueStr) {
         rule.deserialize(valueStr).result().ifPresent(value -> {
             this.initialValues.set(rule, (T)value);
-            this.gameRules.set(rule, (T)value, null);
+            this.gameRules.set(rule, (T)value, (ServerLevel) null);
         });
     }
 
