--- a/net/minecraft/client/gui/screens/options/OnlineOptionsScreen.java
+++ b/net/minecraft/client/gui/screens/options/OnlineOptionsScreen.java
@@ -24,12 +_,9 @@
 @OnlyIn(Dist.CLIENT)
 public class OnlineOptionsScreen extends OptionsSubScreen {
     private static final Component TITLE = Component.translatable("options.online.title");
-    private static final Component SERVERS_HEADER = Component.translatable("options.online.servers.header")
-        .withStyle(ChatFormatting.UNDERLINE, ChatFormatting.BOLD);
-    private static final Component REALMS_HEADER = Component.translatable("options.online.realms.header")
-        .withStyle(ChatFormatting.UNDERLINE, ChatFormatting.BOLD);
-    private static final Component FRIENDS_HEADER = Component.translatable("options.online.friends.header")
-        .withStyle(ChatFormatting.UNDERLINE, ChatFormatting.BOLD);
+    private static final Component SERVERS_HEADER = Component.translatable("options.online.servers.header");
+    private static final Component REALMS_HEADER = Component.translatable("options.online.realms.header");
+    private static final Component FRIENDS_HEADER = Component.translatable("options.online.friends.header");
     private static final Component XBOX_SETTINGS = Component.translatable("options.online.xboxSettings");
     private static final Component FRIENDS_CONFIRM_TITLE = Component.translatable("options.friendsList.confirm.title").withStyle(ChatFormatting.UNDERLINE);
     private static final Component MICROSOFT_ACCOUNT_LINK = Component.translatable("options.friendsList.confirm.message.link")
