--- a/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
@@ -149,6 +_,7 @@
                 orientation = this.orientation.withFront(direction);
             }
 
+            NeighborUpdater.setSourcePos(this.sourcePos); // Youer // Paper - Add source block to BlockPhysicsEvent
             NeighborUpdater.executeUpdate(level, state, neighborPos, this.sourceBlock, orientation, false);
             if (this.idx < NeighborUpdater.UPDATE_ORDER.length && NeighborUpdater.UPDATE_ORDER[this.idx] == this.skipDirection) {
                 this.idx++;
