--- a/net/minecraft/world/level/redstone/ExperimentalRedstoneUtils.java
+++ b/net/minecraft/world/level/redstone/ExperimentalRedstoneUtils.java
@@ -16,6 +_,11 @@
             if (front != null) {
                 orientation = orientation.withFront(front);
             }
+            // Paper start - Optimize redstone (Alternate Current) - use default front instead of random
+            else if (level.paperConfig().misc.redstoneImplementation == io.papermc.paper.configuration.WorldConfiguration.Misc.RedstoneImplementation.ALTERNATE_CURRENT) {
+                orientation = orientation.withFront(Direction.WEST);
+            }
+            // Paper end - Optimize redstone (Alternate Current)
 
             return orientation;
         } else {
