--- a/net/minecraft/world/level/redstone/RedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/RedstoneWireEvaluator.java
@@ -3,14 +_,14 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.world.level.Level;
-import net.minecraft.world.level.block.RedstoneWireBlock;
+import net.minecraft.world.level.block.RedStoneWireBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import org.jspecify.annotations.Nullable;
 
 public abstract class RedstoneWireEvaluator {
-    protected final RedstoneWireBlock wireBlock;
+    protected final RedStoneWireBlock wireBlock;
 
-    protected RedstoneWireEvaluator(RedstoneWireBlock wireBlock) {
+    protected RedstoneWireEvaluator(RedStoneWireBlock wireBlock) {
         this.wireBlock = wireBlock;
     }
 
@@ -23,7 +_,7 @@
     }
 
     protected int getWireSignal(BlockPos pos, BlockState state) {
-        return state.is(this.wireBlock) ? state.getValue(RedstoneWireBlock.POWER) : 0;
+        return state.is(this.wireBlock) ? state.getValue(RedStoneWireBlock.POWER) : 0;
     }
 
     protected int getIncomingWireSignal(Level level, BlockPos pos) {
