--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -5,6 +_,7 @@
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.collect.Table;
+import com.mohistmc.youer.bukkit.inventory.YouerModsInventory;
 import com.mojang.logging.LogUtils;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 import it.unimi.dsi.fastutil.ints.IntList;
@@ -20,9 +_,10 @@
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.network.HashedStack;
+import net.minecraft.network.chat.Component;
+import net.minecraft.resources.Identifier;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.Mth;
-import net.minecraft.util.Prediction;
 import net.minecraft.world.Container;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.SlotAccess;
@@ -33,14 +_,15 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.entity.BlockEntity;
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
 public abstract class AbstractContainerMenu {
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final int SLOT_CLICKED_OUTSIDE = -999;
-    public static final int CONTAINER_CLICK_PRIMARY = 0;
-    public static final int CONTAINER_CLICK_SECONDARY = 1;
     public static final int QUICKCRAFT_TYPE_CHARITABLE = 0;
     public static final int QUICKCRAFT_TYPE_GREEDY = 1;
     public static final int QUICKCRAFT_TYPE_CLONE = 2;
@@ -66,6 +_,56 @@
     private final List<ContainerListener> containerListeners = Lists.newArrayList();
     private @Nullable ContainerSynchronizer synchronizer;
     private boolean suppressRemoteUpdates;
+    @Nullable protected ItemStack activeQuickItem = null; // Purpur - Anvil API
+    // CraftBukkit start
+    public boolean checkReachable = true;
+
+    public InventoryView bukkitView ;
+    public Player containerOwner = null;
+    public InventoryView getBukkitView(){
+        if (bukkitView == null && containerOwner != null) {
+            org.bukkit.inventory.Inventory view = new CraftInventory(new YouerModsInventory(this, containerOwner));
+            bukkitView = new CraftInventoryView<>(containerOwner.getBukkitEntity(), view, this);
+        }
+        return bukkitView;
+    }
+
+    public void transferTo(AbstractContainerMenu other, org.bukkit.craftbukkit.entity.CraftHumanEntity player) {
+        other.containerOwner = player.getHandle();
+        this.containerOwner = player.getHandle();
+        org.bukkit.inventory.InventoryView source = this.getBukkitView(), destination = other.getBukkitView();
+        ((org.bukkit.craftbukkit.inventory.CraftInventory) source.getTopInventory()).getInventory().onClose(player);
+        ((org.bukkit.craftbukkit.inventory.CraftInventory) source.getBottomInventory()).getInventory().onClose(player);
+        ((org.bukkit.craftbukkit.inventory.CraftInventory) destination.getTopInventory()).getInventory().onOpen(player);
+        ((org.bukkit.craftbukkit.inventory.CraftInventory) destination.getBottomInventory()).getInventory().onOpen(player);
+    }
+
+    private net.minecraft.network.chat.@Nullable Component title;
+    public final net.minecraft.network.chat.Component getTitle() {
+        // Paper start - return chat component with empty text instead of throwing error
+        if (this.title == null) {
+            if (this.menuType != null) {
+                Identifier key = BuiltInRegistries.MENU.getKey(this.menuType);
+                if (key == null) {
+                    this.title = Component.literal(this.toString());
+                } else {
+                    this.title = Component.translatable(key.toString());
+                }
+            } else {
+                this.title = Component.literal(this.toString());
+            }
+        }
+        // Paper end - return chat component with empty text instead of throwing error
+        return this.title;
+    }
+
+    public final void setTitle(net.minecraft.network.chat.Component title) {
+        // com.google.common.base.Preconditions.checkState(this.title == null, "Title already set");
+        this.title = title;
+    }
+
+    public void startOpen() {}
+    // CraftBukkit end
 
     protected AbstractContainerMenu(@Nullable MenuType<?> menuType, int containerId) {
         this.menuType = menuType;
@@ -177,8 +_,43 @@
 
         if (this.synchronizer != null) {
             this.synchronizer.sendInitialData(this, itemsToSend, carried.copy(), this.remoteDataSlots.toIntArray());
-        }
-    }
+            this.synchronizer.sendOffHandSlotChange(); // Paper - Sync offhand slot in menus; update player's offhand since the offhand slot is not added to the slots for menus but can be changed by swapping from a menu slot
+        }
+    }
+
+    // Paper start
+    public void forceHeldSlot(final net.minecraft.world.InteractionHand hand) {
+        this.sendAllDataToRemote();
+    }
+
+    public void forceHeldSlotAndArmor(final net.minecraft.world.InteractionHand hand) {
+        this.sendAllDataToRemote();
+    }
+
+    public void forceSlot(final Container container, final int slot) {
+        final int slotsIndex = this.findSlot(container, slot).orElse(-1);
+        if (slotsIndex == -1) {
+            return;
+        }
+
+        final ItemStack item = this.slots.get(slotsIndex).getItem();
+        this.remoteSlots.get(slotsIndex).force(item);
+
+        if (this.synchronizer != null) {
+            this.synchronizer.sendSlotChange(this, slotsIndex, item.copy());
+        }
+    }
+    // Paper end
+
+    // CraftBukkit start
+    public void broadcastCarriedItem() {
+        ItemStack itemstack = this.getCarried().copy();
+        this.remoteCarried.force(itemstack);
+        if (this.synchronizer != null) {
+            this.synchronizer.sendCarriedChange(this, itemstack);
+        }
+    }
+    // CraftBukkit end
 
     public void removeSlotListener(ContainerListener listener) {
         this.containerListeners.remove(listener);
@@ -244,7 +_,7 @@
             this.lastSlots.set(i, newItem);
 
             for (ContainerListener containerListener : this.containerListeners) {
-                containerListener.slotChanged(this, i, newItem);
+                containerListener.slotChanged(this, i, localExpected, newItem); // Paper - Add PlayerInventorySlotChangeEvent
             }
         }
     }
@@ -352,6 +_,7 @@
                     this.resetQuickCraft();
                 }
             } else if (this.quickcraftStatus == 1) {
+                if (slotIndex < 0) return; // Paper - Add slot sanity checks to container clicks
                 Slot slot = this.slots.get(slotIndex);
                 ItemStack carriedItemStack = this.getCarried();
                 if (canItemQuickReplace(slot, carriedItemStack, true)
@@ -376,6 +_,7 @@
                     }
 
                     int remaining = this.getCarried().getCount();
+                    it.unimi.dsi.fastutil.ints.Int2ObjectMap<net.minecraft.world.item.ItemStack> draggedSlots = new it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<>(); // CraftBukkit - Store slots from drag in map (raw slot id -> new stack)
 
                     for (Slot slot : this.quickcraftSlots) {
                         ItemStack carriedItemStack = this.getCarried();
@@ -388,12 +_,42 @@
                             int maxSize = Math.min(source.getMaxStackSize(), slot.getMaxStackSize(source));
                             int newCount = Math.min(getQuickCraftPlaceCount(this.quickcraftSlots.size(), this.quickcraftType, source) + carry, maxSize);
                             remaining -= newCount - carry;
-                            slot.setByPlayer(source.copyWithCount(newCount));
-                        }
-                    }
-
-                    source.setCount(remaining);
-                    this.setCarried(source);
+                            // slot.setByPlayer(source.copyWithCount(newCount));
+                            draggedSlots.put(slot.index, source.copyWithCount(newCount)); // CraftBukkit - Put in map instead of setting
+                        }
+                    }
+
+                    // CraftBukkit start - InventoryDragEvent
+                    org.bukkit.inventory.InventoryView view = this.getBukkitView();
+                    org.bukkit.inventory.ItemStack newCarried = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(source);
+                    newCarried.setAmount(remaining);
+                    java.util.Map<Integer, org.bukkit.inventory.ItemStack> eventMap = new java.util.HashMap<>();
+                    for (it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry<net.minecraft.world.item.ItemStack> entry : draggedSlots.int2ObjectEntrySet()) {
+                        eventMap.put(entry.getIntKey(), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(entry.getValue()));
+                    }
+
+                    // It's essential that we set the cursor to the new value here to prevent item duplication if a plugin closes the inventory.
+                    ItemStack oldCarried = this.getCarried();
+                    this.setCarried(org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(newCarried));
+
+                    org.bukkit.event.inventory.InventoryDragEvent event = new org.bukkit.event.inventory.InventoryDragEvent(
+                            view,
+                            newCarried,
+                            org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(oldCarried),
+                            this.quickcraftType == QUICKCRAFT_TYPE_GREEDY,
+                            eventMap
+                    );
+                    event.callEvent();
+
+                    if (event.getResult() != org.bukkit.event.Event.Result.DENY) {
+                        for (final it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry<net.minecraft.world.item.ItemStack> entry : draggedSlots.int2ObjectEntrySet()) {
+                            view.setItem(entry.getIntKey(), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(entry.getValue()));
+                        }
+                        this.setCarried(org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getCursor()));
+                    } else {
+                        this.setCarried(oldCarried);
+                    }
+                    // CraftBukkit end
                 }
 
                 this.resetQuickCraft();
@@ -402,15 +_,18 @@
             }
         } else if (this.quickcraftStatus != 0) {
             this.resetQuickCraft();
-        } else if ((containerInput == ContainerInput.PICKUP || containerInput == ContainerInput.QUICK_MOVE) && isContainerClickButton(buttonNum)) {
-            ClickAction clickAction = getClickAction(buttonNum);
+        } else if ((containerInput == ContainerInput.PICKUP || containerInput == ContainerInput.QUICK_MOVE) && (buttonNum == 0 || buttonNum == 1)) {
+            ClickAction clickAction = buttonNum == 0 ? ClickAction.PRIMARY : ClickAction.SECONDARY;
             if (slotIndex == -999) {
                 if (!this.getCarried().isEmpty()) {
                     if (clickAction == ClickAction.PRIMARY) {
-                        player.drop(this.getCarried(), true, Prediction.PREDICTED);
+                        // CraftBukkit start
+                        ItemStack carried = this.getCarried();
                         this.setCarried(ItemStack.EMPTY);
+                        player.drop(carried, true);
+                        // CraftBukkit start
                     } else {
-                        player.drop(this.getCarried().split(1), true, Prediction.PREDICTED);
+                        player.drop(this.getCarried().split(1), true);
                     }
                 }
             } else if (containerInput == ContainerInput.QUICK_MOVE) {
@@ -470,8 +_,18 @@
                 }
 
                 slot.setChanged();
+                // CraftBukkit start - Make sure the client has the right slot contents
+                if (player instanceof ServerPlayer serverPlayer && slot.getMaxStackSize() != Container.MAX_STACK) {
+                    serverPlayer.connection.send(new net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket(this.containerId, this.incrementStateId(), slot.index, slot.getItem()));
+                    // Updating a crafting inventory makes the client reset the result slot, have to send it again
+                    if (this.getBukkitView().getType() == org.bukkit.event.inventory.InventoryType.WORKBENCH || this.getBukkitView().getType() == org.bukkit.event.inventory.InventoryType.CRAFTING) {
+                        serverPlayer.connection.send(new net.minecraft.network.protocol.game.ClientboundContainerSetSlotPacket(this.containerId, this.incrementStateId(), 0, this.getSlot(0).getItem()));
+                    }
+                }
+                // CraftBukkit end
             }
         } else if (containerInput == ContainerInput.SWAP && (buttonNum >= 0 && buttonNum < 9 || buttonNum == 40)) {
+            if (slotIndex < 0) return; // Paper - Add slot sanity checks to container clicks
             ItemStack source = inventory.getItem(buttonNum);
             Slot target = this.slots.get(slotIndex);
             ItemStack targetItemStack = target.getItem();
@@ -499,7 +_,7 @@
                         target.setByPlayer(source.split(maxStackSize));
                         target.onTake(player, targetItemStack);
                         if (!inventory.add(targetItemStack)) {
-                            player.drop(targetItemStack, true, Prediction.PREDICTED);
+                            player.drop(targetItemStack, true);
                         }
                     } else {
                         inventory.setItem(buttonNum, targetItemStack);
@@ -521,7 +_,7 @@
             }
 
             ItemStack itemStack = slot.safeTake(amount, Integer.MAX_VALUE, player);
-            player.drop(itemStack, true, Prediction.PREDICTED);
+            player.drop(itemStack, true);
             player.handleCreativeModeItemDrop(itemStack);
             if (buttonNum == 1) {
                 while (!itemStack.isEmpty() && ItemStack.isSameItem(slot.getItem(), itemStack)) {
@@ -530,7 +_,11 @@
                     }
 
                     itemStack = slot.safeTake(amount, Integer.MAX_VALUE, player);
-                    player.drop(itemStack, true, Prediction.PREDICTED);
+                    // CraftBukkit start - SPIGOT-8010: break loop
+                    if (player.drop(itemStack, true) == null) {
+                        break;
+                    }
+                    // CraftBukkit end
                     player.handleCreativeModeItemDrop(itemStack);
                 }
             }
@@ -560,19 +_,13 @@
         }
     }
 
-    private static boolean isContainerClickButton(int buttonNum) {
-        return buttonNum == 0 || buttonNum == 1;
-    }
-
-    private static ClickAction getClickAction(int buttonNum) {
-        return switch (buttonNum) {
-            case 0 -> ClickAction.PRIMARY;
-            case 1 -> ClickAction.SECONDARY;
-            default -> throw new IllegalArgumentException("Not a container click button: " + buttonNum);
-        };
-    }
-
     private boolean tryItemClickBehaviourOverride(Player player, ClickAction clickAction, Slot slot, ItemStack clicked, ItemStack carried) {
+        // Neo: Fire the ItemStackedOnOtherEvent, and return true if it was cancelled (meaning the event was handled). Returning true will trigger the container to stop processing further logic.
+        // The first parameter to onItemStackedOn is the "carried" (under-mouse) item, which is the second ItemStack parameter to this method.
+        if (net.neoforged.neoforge.common.CommonHooks.onItemStackedOn(carried, clicked, slot, clickAction, player, createCarriedSlotAccess())) {
+            return true;
+        }
+
         FeatureFlagSet enabledFeatures = player.level().enabledFeatures();
         return carried.isItemEnabled(enabledFeatures) && carried.overrideStackedOnOther(slot, clickAction, player)
             ? true
@@ -602,20 +_,20 @@
         if (player instanceof ServerPlayer) {
             ItemStack carried = this.getCarried();
             if (!carried.isEmpty()) {
+                this.setCarried(ItemStack.EMPTY); // CraftBukkit - SPIGOT-4556 - from below
                 dropOrPlaceInInventory(player, carried);
-                this.setCarried(ItemStack.EMPTY);
+                // this.setCarried(ItemStack.EMPTY); // CraftBukkit - moved up
             }
         }
     }
 
     private static void dropOrPlaceInInventory(Player player, ItemStack carried) {
-        if (player instanceof ServerPlayer serverPlayer) {
-            boolean playerRemovedNotChangingDimension = serverPlayer.isRemoved() && serverPlayer.getRemovalReason() != Entity.RemovalReason.CHANGED_DIMENSION;
-            if (!playerRemovedNotChangingDimension && !serverPlayer.hasDisconnected()) {
-                serverPlayer.getInventory().placeItemBackInInventory(carried, Prediction.SERVER_ONLY);
-            } else {
-                serverPlayer.drop(carried, false, Prediction.SERVER_ONLY);
-            }
+        boolean playerRemovedNotChangingDimension = !player.isAlive() && player.getRemovalReason() != Entity.RemovalReason.CHANGED_DIMENSION;
+        boolean serverPlayerHasDisconnected = player instanceof ServerPlayer serverPlayer && serverPlayer.hasDisconnected();
+        if (playerRemovedNotChangingDimension || serverPlayerHasDisconnected) {
+            player.drop(carried, false);
+        } else if (player instanceof ServerPlayer) {
+            player.getInventory().placeItemBackInInventory(carried);
         }
     }
 
@@ -650,6 +_,14 @@
     public abstract boolean stillValid(Player player);
 
     protected boolean moveItemStackTo(ItemStack itemStack, int startSlot, int endSlot, boolean backwards) {
+        // Paper start - Add PlayerTradeEvent and PlayerPurchaseEvent
+        return this.moveItemStackTo(itemStack, startSlot, endSlot, backwards, false);
+    }
+    protected boolean moveItemStackTo(ItemStack itemStack, final int startSlot, final int endSlot, final boolean backwards, final boolean isCheck) { // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent - non-final itemStack
+        if (isCheck) {
+            itemStack = itemStack.copy();
+        }
+        // Paper end - Add PlayerTradeEvent and PlayerPurchaseEvent
         boolean anythingChanged = false;
         int destSlot = startSlot;
         if (backwards) {
@@ -660,18 +_,27 @@
             while (!itemStack.isEmpty() && (backwards ? destSlot >= startSlot : destSlot < endSlot)) {
                 Slot slot = this.slots.get(destSlot);
                 ItemStack target = slot.getItem();
+                // Paper start - Add PlayerTradeEvent and PlayerPurchaseEvent; clone if only a check
+                if (isCheck) {
+                    target = target.copy();
+                }
+                // Paper end - Add PlayerTradeEvent and PlayerPurchaseEvent
                 if (!target.isEmpty() && ItemStack.isSameItemSameComponents(itemStack, target)) {
                     int totalStack = target.getCount() + itemStack.getCount();
                     int maxStackSize = slot.getMaxStackSize(target);
                     if (totalStack <= maxStackSize) {
                         itemStack.setCount(0);
                         target.setCount(totalStack);
-                        slot.setChanged();
+                        if (!isCheck) { // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
+                            slot.setChanged();
+                        } // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
                         anythingChanged = true;
                     } else if (target.getCount() < maxStackSize) {
                         itemStack.shrink(maxStackSize - target.getCount());
                         target.setCount(maxStackSize);
-                        slot.setChanged();
+                        if (!isCheck) { // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
+                            slot.setChanged();
+                        } // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
                         anythingChanged = true;
                     }
                 }
@@ -694,10 +_,21 @@
             while (backwards ? destSlot >= startSlot : destSlot < endSlot) {
                 Slot slot = this.slots.get(destSlot);
                 ItemStack target = slot.getItem();
+                // Paper start - Add PlayerTradeEvent and PlayerPurchaseEvent
+                if (isCheck) {
+                    target = target.copy();
+                }
+                // Paper end - Add PlayerTradeEvent and PlayerPurchaseEvent
                 if (target.isEmpty() && slot.mayPlace(itemStack)) {
                     int maxStackSize = slot.getMaxStackSize(itemStack);
-                    slot.setByPlayer(itemStack.split(Math.min(itemStack.getCount(), maxStackSize)));
-                    slot.setChanged();
+                    // Paper start - Add PlayerTradeEvent and PlayerPurchaseEvent
+                    if (isCheck) {
+                        itemStack.shrink(Math.min(itemStack.getCount(), maxStackSize));
+                    } else {
+                        // Paper end - Add PlayerTradeEvent and PlayerPurchaseEvent
+                        slot.setByPlayer(itemStack.split(Math.min(itemStack.getCount(), maxStackSize)));
+                        slot.setChanged();
+                    } // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
                     anythingChanged = true;
                     break;
                 }
@@ -785,6 +_,11 @@
     }
 
     public ItemStack getCarried() {
+        // CraftBukkit start
+        if (this.carried.isEmpty()) {
+            this.setCarried(ItemStack.EMPTY);
+        }
+        // CraftBukkit end
         return this.carried;
     }
 
@@ -838,4 +_,15 @@
         this.stateId = this.stateId + 1 & 32767;
         return this.stateId;
     }
+
+    // Paper start - Add missing InventoryHolders
+    // The reason this is a supplier, is that the createHolder method uses the bukkit InventoryView#getTopInventory to get the inventory in question
+    // and that can't be obtained safely until the AbstractContainerMenu has been fully constructed. Using a supplier lazily
+    // initializes the InventoryHolder safely.
+    protected final Supplier<org.bukkit.inventory.BlockInventoryHolder> createBlockHolder(final ContainerLevelAccess context) {
+        //noinspection ConstantValue
+        com.google.common.base.Preconditions.checkArgument(context != null, "context was null");
+        return () -> context.createBlockHolder(this);
+    }
+    // Paper end - Add missing InventoryHolders
 }
