--- a/net/minecraft/world/inventory/AbstractFurnaceMenu.java
+++ b/net/minecraft/world/inventory/AbstractFurnaceMenu.java
@@ -14,8 +_,12 @@
 import net.minecraft.world.item.crafting.AbstractCookingRecipe;
 import net.minecraft.world.item.crafting.RecipeHolder;
 import net.minecraft.world.item.crafting.RecipePropertySet;
+import net.minecraft.world.item.crafting.RecipeType;
 import net.minecraft.world.item.crafting.SingleRecipeInput;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.AbstractFurnaceBlockEntity;
+import org.bukkit.craftbukkit.inventory.CraftInventoryFurnace;
+import org.bukkit.craftbukkit.inventory.view.CraftFurnaceView;
 
 public abstract class AbstractFurnaceMenu extends RecipeBookMenu {
     public static final int INGREDIENT_SLOT = 0;
@@ -32,6 +_,21 @@
     protected final Level level;
     private final RecipePropertySet acceptedInputs;
     private final RecipeBookType recipeBookType;
+    // CraftBukkit start
+    private org.bukkit.craftbukkit.inventory.view.@org.jspecify.annotations.Nullable CraftFurnaceView view = null;
+    private final Inventory inventory;
+
+    @Override
+    public org.bukkit.craftbukkit.inventory.view.CraftFurnaceView getBukkitView() {
+        if (this.view != null) {
+            return this.view;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventoryFurnace inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryFurnace((net.minecraft.world.level.block.entity.AbstractFurnaceBlockEntity) this.container);
+        this.view = new org.bukkit.craftbukkit.inventory.view.CraftFurnaceView(this.inventory.player.getBukkitEntity(), inventory, this);
+        return this.view;
+    }
+    // CraftBukkit end
 
     protected AbstractFurnaceMenu(
         MenuType<?> menuType, ResourceKey<RecipePropertySet> allowedInputs, RecipeBookType recipeBookType, int containerId, Inventory inventory
@@ -59,6 +_,7 @@
         this.addSlot(new Slot(container, 0, 56, 17));
         this.addSlot(new FurnaceFuelSlot(this, container, 1, 56, 53));
         this.addSlot(new FurnaceResultSlot(inventory.player, container, 2, 116, 35));
+        this.inventory = inventory; // CraftBukkit
         this.addStandardInventorySlots(inventory, 8, 84);
         this.addDataSlots(data);
     }
@@ -76,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.container.stillValid(player);
     }
 
@@ -133,7 +_,13 @@
     }
 
     protected boolean isFuel(ItemStack itemStack) {
-        return this.level.fuelValues().isFuel(itemStack);
+        var recipeType = switch (recipeBookType) {
+            case SMOKER -> net.minecraft.world.item.crafting.RecipeType.SMOKING;
+            case FURNACE -> net.minecraft.world.item.crafting.RecipeType.SMELTING;
+            case BLAST_FURNACE -> net.minecraft.world.item.crafting.RecipeType.BLASTING;
+            default -> null;
+        };
+        return itemStack.getBurnTime(recipeType, this.level.fuelValues()) > 0;
     }
 
     public float getBurnProgress() {
