--- a/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -43,11 +_,11 @@
     public static final int QUICKCRAFT_HEADER_CONTINUE = 1;
     public static final int QUICKCRAFT_HEADER_END = 2;
     public static final int CARRIED_SLOT_SIZE = Integer.MAX_VALUE;
-    public NonNullList<ItemStack> lastSlots = NonNullList.create();
-    public NonNullList<Slot> slots = NonNullList.create();
+    private final NonNullList<ItemStack> lastSlots = NonNullList.create();
+    public final NonNullList<Slot> slots = NonNullList.create();
     private final List<DataSlot> dataSlots = Lists.newArrayList();
     private ItemStack carried = ItemStack.EMPTY;
-    public NonNullList<ItemStack> remoteSlots = NonNullList.create();
+    private final NonNullList<ItemStack> remoteSlots = NonNullList.create();
     private final IntList remoteDataSlots = new IntArrayList();
     private ItemStack remoteCarried = ItemStack.EMPTY;
     private int stateId;
@@ -401,6 +_,7 @@
                 ItemStack itemstack10 = this.getCarried();
                 p_150434_.updateTutorialInventoryAction(itemstack10, slot7.getItem(), clickaction);
                 if (!this.tryItemClickBehaviourOverride(p_150434_, clickaction, slot7, itemstack9, itemstack10)) {
+                if (!net.neoforged.neoforge.common.CommonHooks.onItemStackedOn(itemstack9, itemstack10, slot7, clickaction, p_150434_, createCarriedSlotAccess()))
                     if (itemstack9.isEmpty()) {
                         if (!itemstack10.isEmpty()) {
                             int i3 = clickaction == ClickAction.PRIMARY ? itemstack10.getCount() : 1;
