--- a/net/minecraft/world/inventory/ArmorSlot.java
+++ b/net/minecraft/world/inventory/ArmorSlot.java
@@ -35,7 +_,7 @@
 
     @Override
     public boolean mayPlace(ItemStack itemStack) {
-        return this.owner.isEquippableInSlot(itemStack, this.slot);
+        return itemStack.canEquip(this.slot, this.owner);
     }
 
     @Override
@@ -46,9 +_,8 @@
     @Override
     public boolean mayPickup(Player player) {
         ItemStack itemStack = this.getItem();
-        return !itemStack.isEmpty() && !player.isCreative() && EnchantmentHelper.has(itemStack, EnchantmentEffectComponents.PREVENT_ARMOR_CHANGE)
-            ? false
-            : super.mayPickup(player);
+        return (itemStack.isEmpty() || player.isCreative() || (!EnchantmentHelper.has(itemStack, EnchantmentEffectComponents.PREVENT_ARMOR_CHANGE) || player.level().purpurConfig.playerRemoveBindingWithWeakness && player.hasEffect(net.minecraft.world.effect.MobEffects.WEAKNESS))) // Purpur - Config to remove curse of binding with weakness
+                && super.mayPickup(player);
     }
 
     @Override
