--- a/net/minecraft/world/inventory/BrewingStandMenu.java
+++ b/net/minecraft/world/inventory/BrewingStandMenu.java
@@ -16,6 +_,8 @@
 import net.minecraft.world.item.alchemy.Potion;
 import net.minecraft.world.item.alchemy.PotionBrewing;
 import net.minecraft.world.item.alchemy.PotionContents;
+import org.bukkit.craftbukkit.inventory.CraftInventoryBrewer;
+import org.bukkit.craftbukkit.inventory.view.CraftBrewingStandView;
 
 public class BrewingStandMenu extends AbstractContainerMenu {
     private static final Identifier EMPTY_SLOT_FUEL = Identifier.withDefaultNamespace("container/slot/brewing_fuel");
@@ -34,20 +_,26 @@
     public final ContainerData brewingStandData;
     private final Slot ingredientSlot;
 
+    // CraftBukkit start
+    private CraftBrewingStandView bukkitEntity = null;
+    private Inventory player;
+    // CraftBukkit end
+
     public BrewingStandMenu(int containerId, Inventory inventory) {
         this(containerId, inventory, new SimpleContainer(5), new SimpleContainerData(2));
     }
 
     public BrewingStandMenu(int containerId, Inventory inventory, Container brewingStand, ContainerData brewingStandData) {
         super(MenuType.BREWING_STAND, containerId);
+        player = inventory; // CraftBukkit
         checkContainerSize(brewingStand, 5);
         checkContainerDataCount(brewingStandData, 2);
         this.brewingStand = brewingStand;
         this.brewingStandData = brewingStandData;
         PotionBrewing potionBrewing = inventory.player.level().potionBrewing();
-        this.addSlot(new BrewingStandMenu.PotionSlot(brewingStand, 0, 56, 51));
-        this.addSlot(new BrewingStandMenu.PotionSlot(brewingStand, 1, 79, 58));
-        this.addSlot(new BrewingStandMenu.PotionSlot(brewingStand, 2, 102, 51));
+        this.addSlot(new BrewingStandMenu.PotionSlot(potionBrewing, brewingStand, 0, 56, 51));
+        this.addSlot(new BrewingStandMenu.PotionSlot(potionBrewing, brewingStand, 1, 79, 58));
+        this.addSlot(new BrewingStandMenu.PotionSlot(potionBrewing, brewingStand, 2, 102, 51));
         this.ingredientSlot = this.addSlot(new BrewingStandMenu.IngredientsSlot(potionBrewing, brewingStand, 3, 79, 17));
         this.addSlot(new BrewingStandMenu.FuelSlot(brewingStand, 4, 17, 17));
         this.addDataSlots(brewingStandData);
@@ -56,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.brewingStand.stillValid(player);
     }
 
@@ -75,7 +_,7 @@
                     if (!this.moveItemStackTo(stack, 3, 4, false)) {
                         return ItemStack.EMPTY;
                     }
-                } else if (BrewingStandMenu.PotionSlot.mayPlaceItem(clicked)) {
+                } else if (BrewingStandMenu.PotionSlot.mayPlaceItem(player.level().potionBrewing(), clicked)) {
                     if (!this.moveItemStackTo(stack, 0, 3, false)) {
                         return ItemStack.EMPTY;
                     }
@@ -157,13 +_,20 @@
     }
 
     private static class PotionSlot extends Slot {
+        private final PotionBrewing potionBrewing;
+
         public PotionSlot(Container container, int slot, int x, int y) {
+            this(PotionBrewing.EMPTY, container, slot, x, y);
+        }
+
+        public PotionSlot(PotionBrewing potionBrewing, Container container, int slot, int x, int y) {
             super(container, slot, x, y);
+            this.potionBrewing = potionBrewing;
         }
 
         @Override
         public boolean mayPlace(ItemStack itemStack) {
-            return mayPlaceItem(itemStack);
+            return mayPlaceItem(this.potionBrewing, itemStack);
         }
 
         @Override
@@ -175,19 +_,38 @@
         public void onTake(Player player, ItemStack carried) {
             Optional<Holder<Potion>> potion = carried.getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY).potion();
             if (potion.isPresent() && player instanceof ServerPlayer serverPlayer) {
+                net.neoforged.neoforge.event.EventHooks.onPlayerBrewedPotion(player, carried);
                 CriteriaTriggers.BREWED_POTION.trigger(serverPlayer, potion.get());
             }
 
             super.onTake(player, carried);
         }
 
+        @Deprecated // Neo: use the overload that takes PotionBrewing instead
         public static boolean mayPlaceItem(ItemStack itemStack) {
             return itemStack.is(Items.POTION) || itemStack.is(Items.SPLASH_POTION) || itemStack.is(Items.LINGERING_POTION) || itemStack.is(Items.GLASS_BOTTLE);
         }
 
+        public static boolean mayPlaceItem(PotionBrewing potionBrewing, ItemStack itemStack) {
+            return potionBrewing.isInput(itemStack) || itemStack.is(Items.GLASS_BOTTLE);
+        }
+
         @Override
         public Identifier getNoItemIcon() {
             return BrewingStandMenu.EMPTY_SLOT_POTION;
         }
     }
+
+    // CraftBukkit start
+    @Override
+    public CraftBrewingStandView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryBrewer inventory = new CraftInventoryBrewer(this.brewingStand);
+        bukkitEntity = new CraftBrewingStandView(this.player.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 }
