--- a/net/minecraft/world/inventory/ContainerSynchronizer.java
+++ b/net/minecraft/world/inventory/ContainerSynchronizer.java
@@ -11,4 +_,10 @@
     void sendCarriedChange(AbstractContainerMenu p_150533_, ItemStack p_150534_);
 
     void sendDataChange(AbstractContainerMenu p_150527_, int p_150528_, int p_150529_);
+
+    // Paper start - add flag to simplify remote matching logic
+    default net.minecraft.server.level.@org.jspecify.annotations.Nullable ServerPlayer player() {
+        return null;
+    }
+    // Paper end - add flag to simplify remote matching logic
 }
