--- a/net/minecraft/world/inventory/ContainerLevelAccess.java
+++ b/net/minecraft/world/inventory/ContainerLevelAccess.java
@@ -7,6 +_,21 @@
 import net.minecraft.world.level.Level;
 
 public interface ContainerLevelAccess {
+
+    // CraftBukkit start
+    default Level getWorld() {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    default BlockPos getPosition() {
+        throw new UnsupportedOperationException("Not supported yet.");
+    }
+
+    default org.bukkit.Location getLocation() {
+        return new org.bukkit.Location(getWorld().getWorld(), getPosition().getX(), getPosition().getY(), getPosition().getZ());
+    }
+    // CraftBukkit end
+
     ContainerLevelAccess NULL = new ContainerLevelAccess() {
         @Override
         public <T> Optional<T> evaluate(BiFunction<Level, BlockPos, T> action) {
@@ -16,6 +_,19 @@
 
     static ContainerLevelAccess create(Level level, BlockPos pos) {
         return new ContainerLevelAccess() {
+
+            // CraftBukkit start
+            @Override
+            public Level getWorld() {
+                return level;
+            }
+
+            @Override
+            public BlockPos getPosition() {
+                return pos;
+            }
+            // CraftBukkit end
+
             @Override
             public <T> Optional<T> evaluate(BiFunction<Level, BlockPos, T> action) {
                 return Optional.of(action.apply(level, pos));
