--- a/net/minecraft/world/inventory/DispenserMenu.java
+++ b/net/minecraft/world/inventory/DispenserMenu.java
@@ -12,7 +_,7 @@
     private static final int INV_SLOT_END = 36;
     private static final int USE_ROW_SLOT_START = 36;
     private static final int USE_ROW_SLOT_END = 45;
-    public final Container dispenser;
+    private final Container dispenser;
 
     public DispenserMenu(int p_39433_, Inventory p_39434_) {
         this(p_39433_, p_39434_, new SimpleContainer(9));
