--- a/net/minecraft/world/inventory/EnchantmentMenu.java
+++ b/net/minecraft/world/inventory/EnchantmentMenu.java
@@ -29,6 +_,9 @@
 import net.minecraft.world.item.enchantment.EnchantmentInstance;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.EnchantingTableBlock;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftInventoryEnchanting;
+import org.bukkit.craftbukkit.inventory.view.CraftEnchantmentView;
 
 public class EnchantmentMenu extends AbstractContainerMenu {
     private static final Identifier EMPTY_SLOT_LAPIS_LAZULI = Identifier.withDefaultNamespace("container/slot/lapis_lazuli");
@@ -42,6 +_,13 @@
             super.setChanged();
             EnchantmentMenu.this.slotsChanged(this);
         }
+
+        // CraftBukkit start
+        @Override
+        public Location getLocation() {
+            return access.getLocation();
+        }
+        // CraftBukkit end
     };
     private final ContainerLevelAccess access;
     private final RandomSource random = RandomSource.create();
@@ -49,6 +_,10 @@
     public final int[] costs = new int[3];
     public final int[] enchantClue = new int[]{-1, -1, -1};
     public final int[] levelClue = new int[]{-1, -1, -1};
+    // CraftBukkit start
+    private CraftEnchantmentView bukkitEntity = null;
+    private org.bukkit.entity.Player player;
+    // CraftBukkit end
 
     public EnchantmentMenu(int containerId, Inventory inventory) {
         this(containerId, inventory, ContainerLevelAccess.NULL);
@@ -74,7 +_,7 @@
 
             @Override
             public boolean mayPlace(ItemStack itemStack) {
-                return itemStack.is(Items.LAPIS_LAZULI);
+                return itemStack.is(Items.LAPIS_LAZULI); // Neo: TODO - replace with the tag once we have client tags
             }
 
             @Override
@@ -93,6 +_,9 @@
         this.addDataSlot(DataSlot.shared(this.levelClue, 0));
         this.addDataSlot(DataSlot.shared(this.levelClue, 1));
         this.addDataSlot(DataSlot.shared(this.levelClue, 2));
+        // CraftBukkit start
+        player = (org.bukkit.entity.Player) inventory.player.getBukkitEntity();
+        // CraftBukkit end
     }
 
     @Override
@@ -102,23 +_,24 @@
             if (!itemStack.isEmpty() && itemStack.isEnchantable()) {
                 this.access.execute((level, pos) -> {
                     IdMap<Holder<Enchantment>> holders = level.registryAccess().lookupOrThrow(Registries.ENCHANTMENT).asHolderIdMap();
-                    int bookcases = 0;
+                    float bookcases = 0;
 
                     for (BlockPos offset : EnchantingTableBlock.BOOKSHELF_OFFSETS) {
                         if (EnchantingTableBlock.isValidBookShelf(level, pos, offset)) {
-                            bookcases++;
+                            bookcases += level.getBlockState(pos.offset(offset)).getEnchantPowerBonus(level, pos.offset(offset));
                         }
                     }
 
                     this.random.setSeed(this.enchantmentSeed.get());
 
                     for (int ixx = 0; ixx < 3; ixx++) {
-                        this.costs[ixx] = EnchantmentHelper.getEnchantmentCost(this.random, ixx, bookcases, itemStack);
+                        this.costs[ixx] = EnchantmentHelper.getEnchantmentCost(this.random, ixx, (int) bookcases, itemStack);
                         this.enchantClue[ixx] = -1;
                         this.levelClue[ixx] = -1;
                         if (this.costs[ixx] < ixx + 1) {
                             this.costs[ixx] = 0;
                         }
+                        this.costs[ixx] = net.neoforged.neoforge.event.EventHooks.onEnchantmentLevelSet(level, pos, ixx, (int) bookcases, itemStack, costs[ixx]);
                     }
 
                     for (int ix = 0; ix < 3; ix++) {
@@ -162,14 +_,11 @@
                     List<EnchantmentInstance> newEnchantment = this.getEnchantmentList(level.registryAccess(), itemStack, buttonId, this.costs[buttonId]);
                     if (!newEnchantment.isEmpty()) {
                         player.onEnchantmentPerformed(itemStack, enchantmentCost);
-                        if (itemStack.is(Items.BOOK)) {
-                            enchantmentItem = itemStack.transmuteCopy(Items.ENCHANTED_BOOK);
-                            this.enchantSlots.setItem(0, enchantmentItem);
-                        }
+                        // Neo: Allow items to transform themselves when enchanted, instead of relying on hardcoded transformations for Items.BOOK
+                        enchantmentItem = itemStack.getItem().applyEnchantments(itemStack, newEnchantment);
+                        this.enchantSlots.setItem(0, enchantmentItem);
+                        net.neoforged.neoforge.common.CommonHooks.onPlayerEnchantItem(player, enchantmentItem, newEnchantment);
 
-                        for (EnchantmentInstance enchantment : newEnchantment) {
-                            enchantmentItem.enchant(enchantment.enchantment(), enchantment.level());
-                        }
 
                         currency.consume(enchantmentCost, player);
                         if (currency.isEmpty()) {
@@ -227,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, player, Blocks.ENCHANTING_TABLE);
     }
 
@@ -245,7 +_,7 @@
                 if (!this.moveItemStackTo(stack, 2, 38, true)) {
                     return ItemStack.EMPTY;
                 }
-            } else if (stack.is(Items.LAPIS_LAZULI)) {
+            } else if (stack.is(Items.LAPIS_LAZULI)) { // Neo: TODO - replace with the tag once we have client tags
                 if (!this.moveItemStackTo(stack, 1, 2, true)) {
                     return ItemStack.EMPTY;
                 }
@@ -274,4 +_,17 @@
 
         return clicked;
     }
+
+    // CraftBukkit start
+    @Override
+    public CraftEnchantmentView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryEnchanting inventory = new CraftInventoryEnchanting(this.enchantSlots);
+        bukkitEntity = new CraftEnchantmentView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 }
