--- a/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -20,6 +_,21 @@
 import net.minecraft.world.phys.Vec3;
 
 public class GrindstoneMenu extends AbstractContainerMenu {
+    // CraftBukkit start
+    private org.bukkit.craftbukkit.inventory.@org.jspecify.annotations.Nullable CraftInventoryView view = null;
+    private final org.bukkit.entity.Player player;
+
+    @Override
+    public org.bukkit.craftbukkit.inventory.CraftInventoryView getBukkitView() {
+        if (this.view != null) {
+            return this.view;
+        }
+
+        org.bukkit.craftbukkit.inventory.CraftInventoryGrindstone inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryGrindstone(this.repairSlots, this.resultSlots);
+        this.view = new org.bukkit.craftbukkit.inventory.CraftInventoryView(this.player, inventory, this);
+        return this.view;
+    }
+    // CraftBukkit end
     public static final int MAX_NAME_LENGTH = 35;
     public static final int INPUT_SLOT = 0;
     public static final int ADDITIONAL_SLOT = 1;
@@ -28,15 +_,10 @@
     private static final int INV_SLOT_END = 30;
     private static final int USE_ROW_SLOT_START = 30;
     private static final int USE_ROW_SLOT_END = 39;
-    private final Container resultSlots = new ResultContainer();
-    private final Container repairSlots = new SimpleContainer(2) {
-        @Override
-        public void setChanged() {
-            super.setChanged();
-            GrindstoneMenu.this.slotsChanged(this);
-        }
-    };
+    private final Container resultSlots; // Paper - Add missing InventoryHolders - move down
+    final Container repairSlots; // Paper - Add missing InventoryHolders - move down
     private final ContainerLevelAccess access;
+    private int xp = -1;
 
     public GrindstoneMenu(int containerId, Inventory inventory) {
         this(containerId, inventory, ContainerLevelAccess.NULL);
@@ -44,17 +_,33 @@
 
     public GrindstoneMenu(int containerId, Inventory inventory, ContainerLevelAccess access) {
         super(MenuType.GRINDSTONE, containerId);
+        // Paper start - Add missing InventoryHolders
+        this.resultSlots = new ResultContainer(this.createBlockHolder(access)); // Paper - Add missing InventoryHolders
+        this.repairSlots = new SimpleContainer(this.createBlockHolder(access), 2) { // Paper - Add missing InventoryHolders
+            @Override
+            public void setChanged() {
+                super.setChanged();
+                GrindstoneMenu.this.slotsChanged(this);
+            }
+            // CraftBukkit start
+            @Override
+            public org.bukkit.Location getLocation() {
+                return access.getLocation();
+            }
+            // CraftBukkit end
+        };
+        // Paper end - Add missing InventoryHolders
         this.access = access;
         this.addSlot(new Slot(this.repairSlots, 0, 49, 19) {
             @Override
             public boolean mayPlace(ItemStack itemStack) {
-                return itemStack.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(itemStack);
+                return itemStack.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(itemStack) || itemStack.canGrindstoneRepair();
             }
         });
         this.addSlot(new Slot(this.repairSlots, 1, 49, 40) {
             @Override
             public boolean mayPlace(ItemStack itemStack) {
-                return itemStack.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(itemStack);
+                return itemStack.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(itemStack) || itemStack.canGrindstoneRepair();
             }
         });
         this.addSlot(new Slot(this.resultSlots, 2, 129, 34) {
@@ -65,9 +_,16 @@
 
             @Override
             public void onTake(Player player, ItemStack carried) {
+                if (net.neoforged.neoforge.common.CommonHooks.onGrindstoneTake(GrindstoneMenu.this.repairSlots, access, player, this::getExperienceAmount)) return;
                 access.execute((level, pos) -> {
+                    ItemStack itemstack = activeQuickItem == null ? carried : activeQuickItem; // Purpur - Grindstone API
                     if (level instanceof ServerLevel serverLevel) {
-                        ExperienceOrb.award(serverLevel, Vec3.atCenterOf(pos), this.getExperienceAmount(level));
+                        // Paper start - Fire BlockExpEvent on grindstone use
+                        org.bukkit.event.block.BlockExpEvent event = new org.bukkit.event.block.BlockExpEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), this.getExperienceAmount(level));
+                        event.callEvent();
+                        org.purpurmc.purpur.event.inventory.GrindstoneTakeResultEvent grindstoneTakeResultEvent = new org.purpurmc.purpur.event.inventory.GrindstoneTakeResultEvent(player.getBukkitEntity(), getBukkitView(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), event.getExpToDrop()); grindstoneTakeResultEvent.callEvent(); // Purpur - Grindstone API
+                        ExperienceOrb.awardWithDirection(serverLevel, Vec3.atCenterOf(pos), Vec3.ZERO, grindstoneTakeResultEvent.getExperienceAmount(), org.bukkit.entity.ExperienceOrb.SpawnReason.GRINDSTONE, player, null); // Purpur - Grindstone API
+                        // Paper end - Fire BlockExpEvent on grindstone use
                     }
 
                     level.levelEvent(1042, pos, 0);
@@ -77,6 +_,7 @@
             }
 
             private int getExperienceAmount(Level level) {
+                if (xp > -1) return xp;
                 int amount = 0;
                 amount += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(0));
                 amount += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(1));
@@ -95,7 +_,7 @@
                 for (Entry<Holder<Enchantment>> entry : enchantments.entrySet()) {
                     Holder<Enchantment> enchant = entry.getKey();
                     int lvl = entry.getIntValue();
-                    if (!enchant.is(EnchantmentTags.CURSE)) {
+                    if (!org.purpurmc.purpur.PurpurConfig.grindstoneIgnoredEnchants.contains(enchant.value())) { // Purpur - Config for grindstones
                         amount += enchant.value().getMinCost(lvl);
                     }
                 }
@@ -104,6 +_,7 @@
             }
         });
         this.addStandardInventorySlots(inventory, 8, 84);
+        player = (org.bukkit.entity.Player) inventory.player.getBukkitEntity(); // CraftBukkit
     }
 
     @Override
@@ -111,11 +_,15 @@
         super.slotsChanged(container);
         if (container == this.repairSlots) {
             this.createResult();
+            org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareResultEvent(this, RESULT_SLOT); // Paper - Add PrepareResultEvent
         }
     }
 
     private void createResult() {
-        this.resultSlots.setItem(0, this.computeResult(this.repairSlots.getItem(0), this.repairSlots.getItem(1)));
+        this.xp = net.neoforged.neoforge.common.CommonHooks.onGrindstoneChange(this.repairSlots.getItem(0), this.repairSlots.getItem(1), this.resultSlots, -1);
+        if (this.xp == Integer.MIN_VALUE)
+        org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareGrindstoneEvent(this.getBukkitView(), this.computeResult(this.repairSlots.getItem(0), this.repairSlots.getItem(1))); // CraftBukkit
+        this.sendAllDataToRemote(); // CraftBukkit - SPIGOT-6686: Always send completed inventory to stay in sync with client
         this.broadcastChanges();
     }
 
@@ -148,7 +_,7 @@
         int remaining2 = additional.getMaxDamage() - additional.getDamageValue();
         int remaining = remaining1 + remaining2 + durability * 5 / 100;
         int count = 1;
-        if (!input.isDamageableItem()) {
+        if (!input.isDamageableItem()  || !input.isCombineRepairable()) {
             if (input.getMaxStackSize() < 2 || !ItemStack.matches(input, additional)) {
                 return ItemStack.EMPTY;
             }
@@ -160,6 +_,7 @@
         if (newItem.isDamageableItem()) {
             newItem.set(DataComponents.MAX_DAMAGE, durability);
             newItem.setDamageValue(Math.max(durability - remaining, 0));
+            if (!additional.isCombineRepairable()) newItem.setDamageValue(input.getDamageValue());
         }
 
         this.mergeEnchantsFrom(newItem, additional);
@@ -172,16 +_,76 @@
 
             for (Entry<Holder<Enchantment>> entry : enchantments.entrySet()) {
                 Holder<Enchantment> enchant = entry.getKey();
-                if (!enchant.is(EnchantmentTags.CURSE) || newEnchantments.getLevel(enchant) == 0) {
+                if (!org.purpurmc.purpur.PurpurConfig.grindstoneIgnoredEnchants.contains(enchant.value()) || newEnchantments.getLevel(enchant) == 0) { // Purpur - Config for grindstones
                     newEnchantments.upgrade(enchant, entry.getIntValue());
                 }
             }
         });
     }
 
+    // Purpur start - Config for grindstones
+    private java.util.List<net.minecraft.core.component.DataComponentType<?>> GRINDSTONE_REMOVE_ATTRIBUTES_REMOVAL_LIST = java.util.List.of(
+            // DataComponents.MAX_STACK_SIZE,
+            // DataComponents.DAMAGE,
+            // DataComponents.BLOCK_STATE,
+            DataComponents.CUSTOM_DATA,
+            // DataComponents.MAX_DAMAGE,
+            // DataComponents.UNBREAKABLE,
+            // DataComponents.CUSTOM_NAME,
+            // DataComponents.ITEM_NAME,
+            // DataComponents.LORE,
+            // DataComponents.RARITY,
+            // DataComponents.ENCHANTMENTS,
+            // DataComponents.CAN_PLACE_ON,
+            // DataComponents.CAN_BREAK,
+            DataComponents.ATTRIBUTE_MODIFIERS,
+            DataComponents.CUSTOM_MODEL_DATA,
+            // DataComponents.HIDE_ADDITIONAL_TOOLTIP,
+            // DataComponents.HIDE_TOOLTIP,
+            // DataComponents.REPAIR_COST,
+            // DataComponents.CREATIVE_SLOT_LOCK,
+            // DataComponents.ENCHANTMENT_GLINT_OVERRIDE,
+            // DataComponents.INTANGIBLE_PROJECTILE,
+            // DataComponents.FOOD,
+            // DataComponents.FIRE_RESISTANT,
+            // DataComponents.TOOL,
+            // DataComponents.STORED_ENCHANTMENTS,
+            DataComponents.DYED_COLOR,
+            // DataComponents.MAP_COLOR,
+            // DataComponents.MAP_ID,
+            // DataComponents.MAP_DECORATIONS,
+            // DataComponents.MAP_POST_PROCESSING,
+            // DataComponents.CHARGED_PROJECTILES,
+            // DataComponents.BUNDLE_CONTENTS,
+            // DataComponents.POTION_CONTENTS,
+            DataComponents.SUSPICIOUS_STEW_EFFECTS
+            // DataComponents.WRITABLE_BOOK_CONTENT,
+            // DataComponents.WRITTEN_BOOK_CONTENT,
+            // DataComponents.TRIM,
+            // DataComponents.DEBUG_STICK_STATE,
+            // DataComponents.ENTITY_DATA,
+            // DataComponents.BUCKET_ENTITY_DATA,
+            // DataComponents.BLOCK_ENTITY_DATA,
+            // DataComponents.INSTRUMENT,
+            // DataComponents.OMINOUS_BOTTLE_AMPLIFIER,
+            // DataComponents.RECIPES,
+            // DataComponents.LODESTONE_TRACKER,
+            // DataComponents.FIREWORK_EXPLOSION,
+            // DataComponents.FIREWORKS,
+            // DataComponents.PROFILE,
+            // DataComponents.NOTE_BLOCK_SOUND,
+            // DataComponents.BANNER_PATTERNS,
+            // DataComponents.BASE_COLOR,
+            // DataComponents.POT_DECORATIONS,
+            // DataComponents.CONTAINER,
+            // DataComponents.BEES,
+            // DataComponents.LOCK,
+            // DataComponents.CONTAINER_LOOT,
+    );
+    // Purpur end - Config for grindstones
     private ItemStack removeNonCursesFrom(ItemStack item) {
         ItemEnchantments newEnchantments = EnchantmentHelper.updateEnchantments(
-            item, enchantments -> enchantments.removeIf(enchantment -> !enchantment.is(EnchantmentTags.CURSE))
+            item, enchantments -> enchantments.removeIf(enchantment -> !org.purpurmc.purpur.PurpurConfig.grindstoneIgnoredEnchants.contains(enchantment.value())) // Purpur - Config for grindstones
         );
         if (item.is(Items.ENCHANTED_BOOK) && newEnchantments.isEmpty()) {
             item = item.transmuteCopy(Items.BOOK);
@@ -194,6 +_,22 @@
         }
 
         item.set(DataComponents.REPAIR_COST, repairCost);
+
+        // Purpur start - Config for grindstones
+        net.minecraft.core.component.DataComponentPatch.Builder builder = net.minecraft.core.component.DataComponentPatch.builder();
+        if (org.purpurmc.purpur.PurpurConfig.grindstoneRemoveAttributes) {
+            item.getComponents().forEach(typedDataComponent -> {
+                if (GRINDSTONE_REMOVE_ATTRIBUTES_REMOVAL_LIST.contains(typedDataComponent.type())) {
+                    builder.remove(typedDataComponent.type());
+                }
+            });
+        }
+        if (org.purpurmc.purpur.PurpurConfig.grindstoneRemoveDisplay) {
+            builder.remove(DataComponents.CUSTOM_NAME);
+            builder.remove(DataComponents.LORE);
+        }
+        item.applyComponents(builder.build());
+        // Purpur end - Config for grindstones
         return item;
     }
 
@@ -205,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, player, Blocks.GRINDSTONE);
     }
 
@@ -249,7 +_,9 @@
                 return ItemStack.EMPTY;
             }
 
+            this.activeQuickItem = clicked; // Purpur - Grindstone API
             slot.onTake(player, item);
+            this.activeQuickItem = null; // Purpur - Grindstone API
         }
 
         return clicked;
