--- a/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -19,8 +_,27 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftInventoryGrindstone;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class GrindstoneMenu extends AbstractContainerMenu {
+
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private org.bukkit.entity.Player player;
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryGrindstone inventory = new CraftInventoryGrindstone(this.repairSlots, this.resultSlots);
+        bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
     public static final int MAX_NAME_LENGTH = 35;
     public static final int INPUT_SLOT = 0;
     public static final int ADDITIONAL_SLOT = 1;
@@ -40,8 +_,16 @@
             super.setChanged();
             GrindstoneMenu.this.slotsChanged(this);
         }
+
+        // CraftBukkit start
+        @Override
+        public Location getLocation() {
+            return access.getLocation();
+        }
+        // CraftBukkit end
     };
     private final ContainerLevelAccess access;
+    private int xp = -1;
 
     public GrindstoneMenu(int containerId, Inventory inventory) {
         this(containerId, inventory, ContainerLevelAccess.NULL);
@@ -57,7 +_,7 @@
 
             @Override
             public boolean mayPlace(ItemStack itemStack) {
-                return itemStack.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(itemStack);
+                return itemStack.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(itemStack) || itemStack.canGrindstoneRepair();
             }
         });
         this.addSlot(new Slot(this.repairSlots, 1, 49, 40) {
@@ -67,7 +_,7 @@
 
             @Override
             public boolean mayPlace(ItemStack itemStack) {
-                return itemStack.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(itemStack);
+                return itemStack.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(itemStack) || itemStack.canGrindstoneRepair();
             }
         });
         this.addSlot(new Slot(this.resultSlots, 2, 129, 34) {
@@ -82,6 +_,7 @@
 
             @Override
             public void onTake(Player player, ItemStack carried) {
+                if (net.neoforged.neoforge.common.CommonHooks.onGrindstoneTake(GrindstoneMenu.this.repairSlots, access, player, this::getExperienceAmount)) return;
                 access.execute((level, pos) -> {
                     if (level instanceof ServerLevel) {
                         ExperienceOrb.award((ServerLevel)level, Vec3.atCenterOf(pos), this.getExperienceAmount(level));
@@ -94,6 +_,7 @@
             }
 
             private int getExperienceAmount(Level level) {
+                if (xp > -1) return xp;
                 int amount = 0;
                 amount += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(0));
                 amount += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(1));
@@ -121,6 +_,7 @@
             }
         });
         this.addStandardInventorySlots(inventory, 8, 84);
+        player = (org.bukkit.entity.Player) inventory.player.getBukkitEntity(); // CraftBukkit
     }
 
     @Override
@@ -132,7 +_,10 @@
     }
 
     private void createResult() {
-        this.resultSlots.setItem(0, this.computeResult(this.repairSlots.getItem(0), this.repairSlots.getItem(1)));
+        this.xp = net.neoforged.neoforge.common.CommonHooks.onGrindstoneChange(this.repairSlots.getItem(0), this.repairSlots.getItem(1), this.resultSlots, -1);
+        if (this.xp == Integer.MIN_VALUE)
+        org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareGrindstoneEvent(getBukkitView(), this.computeResult(this.repairSlots.getItem(0), this.repairSlots.getItem(1))); // CraftBukkit
+        sendAllDataToRemote(); // CraftBukkit - SPIGOT-6686: Always send completed inventory to stay in sync with client
         this.broadcastChanges();
     }
 
@@ -162,7 +_,7 @@
             int remaining2 = additional.getMaxDamage() - additional.getDamageValue();
             int remaining = remaining1 + remaining2 + durability * 5 / 100;
             int count = 1;
-            if (!input.isDamageableItem()) {
+            if (!input.isDamageableItem() || !input.isCombineRepairable()) {
                 if (input.getMaxStackSize() < 2 || !ItemStack.matches(input, additional)) {
                     return ItemStack.EMPTY;
                 }
@@ -174,6 +_,7 @@
             if (newItem.isDamageableItem()) {
                 newItem.set(DataComponents.MAX_DAMAGE, durability);
                 newItem.setDamageValue(Math.max(durability - remaining, 0));
+                if (!additional.isCombineRepairable()) newItem.setDamageValue(input.getDamageValue());
             }
 
             this.mergeEnchantsFrom(newItem, additional);
@@ -220,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, player, Blocks.GRINDSTONE);
     }
 
