--- a/net/minecraft/world/inventory/GrindstoneMenu.java
+++ b/net/minecraft/world/inventory/GrindstoneMenu.java
@@ -1,7 +_,6 @@
 package net.minecraft.world.inventory;
 
 import it.unimi.dsi.fastutil.objects.Object2IntMap.Entry;
-import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.server.level.ServerLevel;
@@ -19,6 +_,10 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftInventoryGrindstone;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
 
 public class GrindstoneMenu extends AbstractContainerMenu {
     public static final int MAX_NAME_LENGTH = 35;
@@ -36,8 +_,32 @@
             super.setChanged();
             GrindstoneMenu.this.slotsChanged(this);
         }
+
+        // CraftBukkit start
+        @Override
+        public Location getLocation() {
+            return access.getLocation();
+        }
+            // CraftBukkit end
     };
     private final ContainerLevelAccess access;
+    private int xp = -1;
+
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private org.bukkit.entity.Player player;
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryGrindstone inventory = new CraftInventoryGrindstone(this.repairSlots, this.resultSlots);
+        bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 
     public GrindstoneMenu(int p_39563_, Inventory p_39564_) {
         this(p_39563_, p_39564_, ContainerLevelAccess.NULL);
@@ -49,13 +_,13 @@
         this.addSlot(new Slot(this.repairSlots, 0, 49, 19) {
             @Override
             public boolean mayPlace(ItemStack p_39607_) {
-                return p_39607_.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(p_39607_);
+                return p_39607_.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(p_39607_) || p_39607_.canGrindstoneRepair();
             }
         });
         this.addSlot(new Slot(this.repairSlots, 1, 49, 40) {
             @Override
             public boolean mayPlace(ItemStack p_39616_) {
-                return p_39616_.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(p_39616_);
+                return p_39616_.isDamageableItem() || EnchantmentHelper.hasAnyEnchantments(p_39616_) || p_39616_.canGrindstoneRepair();
             }
         });
         this.addSlot(new Slot(this.resultSlots, 2, 129, 34) {
@@ -66,9 +_,10 @@
 
             @Override
             public void onTake(Player p_150574_, ItemStack p_150575_) {
+                if (net.neoforged.neoforge.common.CommonHooks.onGrindstoneTake(GrindstoneMenu.this.repairSlots, p_39568_, p_150574_, this::getExperienceAmount)) return;
                 p_39568_.execute((p_39634_, p_39635_) -> {
                     if (p_39634_ instanceof ServerLevel) {
-                        ExperienceOrb.award((ServerLevel)p_39634_, Vec3.atCenterOf(p_39635_), this.getExperienceAmount(p_39634_));
+                        ExperienceOrb.award((ServerLevel)p_39634_, Vec3.atCenterOf(p_39635_), this.getExperienceAmount(p_39634_), org.bukkit.entity.ExperienceOrb.SpawnReason.GRINDSTONE, p_150574_); // Paper
                     }
 
                     p_39634_.levelEvent(1042, p_39635_, 0);
@@ -78,6 +_,7 @@
             }
 
             private int getExperienceAmount(Level p_39632_) {
+                if (xp > -1) return xp;
                 int l = 0;
                 l += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(0));
                 l += this.getExperienceFromItem(GrindstoneMenu.this.repairSlots.getItem(1));
@@ -114,6 +_,8 @@
         for (int k = 0; k < 9; k++) {
             this.addSlot(new Slot(p_39567_, k, 8 + k * 18, 142));
         }
+
+        this.player = (org.bukkit.entity.Player) p_39567_.player.getBukkitEntity(); // CraftBukkit
     }
 
     @Override
@@ -121,11 +_,16 @@
         super.slotsChanged(p_39570_);
         if (p_39570_ == this.repairSlots) {
             this.createResult();
+            org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareResultEvent(this, RESULT_SLOT); // Paper - Add PrepareResultEvent
         }
     }
 
     private void createResult() {
-        this.resultSlots.setItem(0, this.computeResult(this.repairSlots.getItem(0), this.repairSlots.getItem(1)));
+        this.xp = net.neoforged.neoforge.common.CommonHooks.onGrindstoneChange(this.repairSlots.getItem(0), this.repairSlots.getItem(1), this.resultSlots, -1);
+        if (this.xp == Integer.MIN_VALUE)
+        CraftEventFactory.callPrepareGrindstoneEvent(getBukkitView(), this.computeResult(this.repairSlots.getItem(0), this.repairSlots.getItem(1))); // CraftBukkit
+        sendAllDataToRemote(); // CraftBukkit - SPIGOT-6686: Always send completed inventory to stay in sync with client
+
         this.broadcastChanges();
     }
 
@@ -155,7 +_,7 @@
             int k = p_332686_.getMaxDamage() - p_332686_.getDamageValue();
             int l = j + k + i * 5 / 100;
             int i1 = 1;
-            if (!p_332723_.isDamageableItem()) {
+            if (!p_332723_.isDamageableItem() || !p_332723_.isRepairable()) {
                 if (p_332723_.getMaxStackSize() < 2 || !ItemStack.matches(p_332723_, p_332686_)) {
                     return ItemStack.EMPTY;
                 }
@@ -167,6 +_,7 @@
             if (itemstack.isDamageableItem()) {
                 itemstack.set(DataComponents.MAX_DAMAGE, i);
                 itemstack.setDamageValue(Math.max(i - l, 0));
+                if (!p_332686_.isRepairable()) itemstack.setDamageValue(p_332723_.getDamageValue());
             }
 
             this.mergeEnchantsFrom(itemstack, p_332686_);
@@ -213,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player p_39572_) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, p_39572_, Blocks.GRINDSTONE);
     }
 
