--- a/net/minecraft/world/inventory/ItemCombinerMenu.java
+++ b/net/minecraft/world/inventory/ItemCombinerMenu.java
@@ -50,15 +_,15 @@
     }
 
     private void createInputSlots(ItemCombinerMenuSlotDefinition itemInputSlots) {
-        for (final ItemCombinerMenuSlotDefinition.SlotDefinition slot : itemInputSlots.getSlots()) {
-            this.addSlot(new Slot(this.inputSlots, slot.slotIndex(), slot.x(), slot.y()) {
+        for (final ItemCombinerMenuSlotDefinition.SlotDefinition slot1 : itemInputSlots.getSlots()) {
+            this.addSlot(new Slot(this.inputSlots, slot1.slotIndex(), slot1.x(), slot1.y()) {
                 {
                     Objects.requireNonNull(ItemCombinerMenu.this);
                 }
 
                 @Override
                 public boolean mayPlace(ItemStack itemStack) {
-                    return slot.mayPlace().test(itemStack);
+                    return slot1.mayPlace().test(itemStack);
                 }
             });
         }
@@ -121,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.access
             .evaluate((level, pos) -> !this.isValidBlock(level.getBlockState(pos)) ? false : player.isWithinBlockInteractionRange(pos, 4.0), true);
     }
