--- a/net/minecraft/world/inventory/LoomMenu.java
+++ b/net/minecraft/world/inventory/LoomMenu.java
@@ -2,6 +_,7 @@
 
 import com.google.common.collect.ImmutableList;
 import java.util.List;
+import java.util.Objects;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderSet;
@@ -21,8 +_,27 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.BannerPattern;
 import net.minecraft.world.level.block.entity.BannerPatternLayers;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftInventoryLoom;
+import org.bukkit.craftbukkit.inventory.view.CraftLoomView;
 
 public class LoomMenu extends AbstractContainerMenu {
+
+    // CraftBukkit start
+    private CraftLoomView bukkitEntity = null;
+    private org.bukkit.entity.Player player;
+
+    @Override
+    public CraftLoomView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryLoom inventory = new CraftInventoryLoom(this.inputContainer, this.outputContainer);
+        bukkitEntity = new CraftLoomView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
     private static final int PATTERN_NOT_SET = -1;
     private static final int INV_SLOT_START = 4;
     private static final int INV_SLOT_END = 31;
@@ -45,6 +_,13 @@
             LoomMenu.this.slotsChanged(this);
             LoomMenu.this.slotUpdateListener.run();
         }
+
+        // CraftBukkit start
+        @Override
+        public Location getLocation() {
+            return access.getLocation();
+        }
+        // CraftBukkit end
     };
     private final Container outputContainer = new SimpleContainer(1) {
         @Override
@@ -52,6 +_,13 @@
             super.setChanged();
             LoomMenu.this.slotUpdateListener.run();
         }
+
+        // CraftBukkit start
+        @Override
+        public Location getLocation() {
+            return access.getLocation();
+        }
+        // CraftBukkit end
     };
 
     public LoomMenu(int containerId, Inventory inventory) {
@@ -106,6 +_,7 @@
         this.addStandardInventorySlots(inventory, 8, 84);
         this.addDataSlot(this.selectedBannerPatternIndex);
         this.patternGetter = inventory.player.registryAccess().lookupOrThrow(Registries.BANNER_PATTERN);
+        player = (org.bukkit.entity.Player) inventory.player.getBukkitEntity(); // CraftBukkit
     }
 
     private static boolean isPatternItem(ItemStack itemStack) {
@@ -118,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, player, Blocks.LOOM);
     }
 
@@ -277,7 +_,14 @@
                 result.update(
                     DataComponents.BANNER_PATTERNS,
                     BannerPatternLayers.EMPTY,
-                    layers -> new BannerPatternLayers.Builder().addAll(layers).add(pattern, patternColor).build()
+                    layers -> {
+                        // CraftBukkit start
+                        if (layers.layers().size() > 20) {
+                            layers = new BannerPatternLayers(List.copyOf(layers.layers().subList(0, 20)));
+                        }
+                        // CraftBukkit end
+                        return new BannerPatternLayers.Builder().addAll(layers).add(pattern, patternColor).build();
+                    }
                 );
             }
         }
