--- a/net/minecraft/world/inventory/LoomMenu.java
+++ b/net/minecraft/world/inventory/LoomMenu.java
@@ -2,7 +_,6 @@
 
 import com.google.common.collect.ImmutableList;
 import java.util.List;
-import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.component.DataComponents;
@@ -19,10 +_,12 @@
 import net.minecraft.world.item.DyeColor;
 import net.minecraft.world.item.DyeItem;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.BannerPattern;
 import net.minecraft.world.level.block.entity.BannerPatternLayers;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftInventoryLoom;
+import org.bukkit.craftbukkit.inventory.view.CraftLoomView;
 
 public class LoomMenu extends AbstractContainerMenu {
     private static final int PATTERN_NOT_SET = -1;
@@ -48,6 +_,12 @@
             LoomMenu.this.slotsChanged(this);
             LoomMenu.this.slotUpdateListener.run();
         }
+        // CraftBukkit start
+        @Override
+        public Location getLocation() {
+            return access.getLocation();
+        }
+        // CraftBukkit end
     };
     private final Container outputContainer = new SimpleContainer(1) {
         @Override
@@ -55,7 +_,29 @@
             super.setChanged();
             LoomMenu.this.slotUpdateListener.run();
         }
+        // CraftBukkit start
+        @Override
+        public Location getLocation() {
+            return access.getLocation();
+        }
+        // CraftBukkit end
     };
+
+    // CraftBukkit start
+    private CraftLoomView bukkitEntity = null;
+    private org.bukkit.entity.Player player;
+
+    @Override
+    public CraftLoomView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryLoom inventory = new CraftInventoryLoom(this.inputContainer, this.outputContainer);
+        bukkitEntity = new CraftLoomView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 
     public LoomMenu(int p_39856_, Inventory p_39857_) {
         this(p_39856_, p_39857_, ContainerLevelAccess.NULL);
@@ -119,18 +_,44 @@
 
         this.addDataSlot(this.selectedBannerPatternIndex);
         this.patternGetter = p_39860_.player.registryAccess().lookupOrThrow(Registries.BANNER_PATTERN);
+        player = (org.bukkit.entity.Player) p_39860_.player.getBukkitEntity(); // CraftBukkit
     }
 
     @Override
     public boolean stillValid(Player p_39865_) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, p_39865_, Blocks.LOOM);
     }
 
     @Override
     public boolean clickMenuButton(Player p_39867_, int p_39868_) {
         if (p_39868_ >= 0 && p_39868_ < this.selectablePatterns.size()) {
-            this.selectedBannerPatternIndex.set(p_39868_);
-            this.setupResultSlot(this.selectablePatterns.get(p_39868_));
+            // Paper start - Add PlayerLoomPatternSelectEvent
+            int selectablePatternIndex = p_39868_;
+            io.papermc.paper.event.player.PlayerLoomPatternSelectEvent event = new io.papermc.paper.event.player.PlayerLoomPatternSelectEvent((org.bukkit.entity.Player) p_39867_.getBukkitEntity(), ((CraftInventoryLoom) getBukkitView().getTopInventory()), org.bukkit.craftbukkit.block.banner.CraftPatternType.minecraftHolderToBukkit(this.selectablePatterns.get(selectablePatternIndex)));
+            if (!event.callEvent()) {
+                p_39867_.containerMenu.sendAllDataToRemote();
+                return false;
+            }
+            final Holder<BannerPattern> eventPattern = org.bukkit.craftbukkit.block.banner.CraftPatternType.bukkitToMinecraftHolder(event.getPatternType());
+            Holder<BannerPattern> selectedPattern = null;
+            for (int i = 0; i < this.selectablePatterns.size(); i++) {
+                final Holder<BannerPattern> holder = this.selectablePatterns.get(i);
+                if (eventPattern.equals(holder)) {
+                    selectablePatternIndex = i;
+                    selectedPattern = holder;
+                    break;
+                }
+            }
+            if (selectedPattern == null) {
+                selectedPattern = eventPattern;
+                selectablePatternIndex = -1;
+            }
+
+            p_39867_.containerMenu.sendAllDataToRemote();
+            this.selectedBannerPatternIndex.set(selectablePatternIndex);
+            this.setupResultSlot(java.util.Objects.requireNonNull(selectedPattern, "selectedPattern was null, this is unexpected"));
+            // Paper end - Add PlayerLoomPatternSelectEvent
             return true;
         } else {
             return false;
@@ -193,7 +_,7 @@
                 this.resultSlot.set(ItemStack.EMPTY);
             }
 
-            this.broadcastChanges();
+            org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareResultEvent(this, 3); // Paper - Add PrepareResultEvent
         } else {
             this.resultSlot.set(ItemStack.EMPTY);
             this.selectablePatterns = List.of();
@@ -282,7 +_,14 @@
             itemstack2.update(
                 DataComponents.BANNER_PATTERNS,
                 BannerPatternLayers.EMPTY,
-                p_330070_ -> new BannerPatternLayers.Builder().addAll(p_330070_).add(p_219992_, dyecolor).build()
+                p_330070_ -> {
+                    // CraftBukkit start
+                    if (p_330070_.layers().size() > 20) {
+                        p_330070_ = new BannerPatternLayers(List.copyOf(p_330070_.layers().subList(0, 20)));
+                    }
+                    // CraftBukkit end
+                    return new BannerPatternLayers.Builder().addAll(p_330070_).add(p_219992_, dyecolor).build();
+                }
             );
         }
 
