--- a/net/minecraft/world/inventory/MerchantContainer.java
+++ b/net/minecraft/world/inventory/MerchantContainer.java
@@ -15,7 +_,7 @@
     private final NonNullList<ItemStack> itemStacks = NonNullList.withSize(3, ItemStack.EMPTY);
     @Nullable
     private MerchantOffer activeOffer;
-    public int selectionHint;
+    private int selectionHint;
     private int futureXp;
 
     public MerchantContainer(Merchant p_40003_) {
