--- a/net/minecraft/world/inventory/MerchantMenu.java
+++ b/net/minecraft/world/inventory/MerchantMenu.java
@@ -12,6 +_,8 @@
 import net.minecraft.world.item.trading.Merchant;
 import net.minecraft.world.item.trading.MerchantOffer;
 import net.minecraft.world.item.trading.MerchantOffers;
+import org.bukkit.craftbukkit.inventory.CraftInventoryMerchant;
+import org.bukkit.craftbukkit.inventory.view.CraftMerchantView;
 
 public class MerchantMenu extends AbstractContainerMenu {
     protected static final int PAYMENT1_SLOT = 0;
@@ -31,6 +_,19 @@
     private boolean showProgressBar;
     private boolean canRestock;
 
+    // CraftBukkit start
+    private CraftMerchantView bukkitEntity = null;
+    private Inventory player;
+
+    @Override
+    public CraftMerchantView getBukkitView() {
+        if (bukkitEntity == null) {
+            bukkitEntity = new CraftMerchantView(this.player.player.getBukkitEntity(), new CraftInventoryMerchant(trader, tradeContainer), this, trader);
+        }
+        return bukkitEntity;
+    }
+    // CraftBukkit end
+
     public MerchantMenu(int p_40033_, Inventory p_40034_) {
         this(p_40033_, p_40034_, new ClientSideMerchant(p_40034_.player));
     }
@@ -42,6 +_,7 @@
         this.addSlot(new Slot(this.tradeContainer, 0, 136, 37));
         this.addSlot(new Slot(this.tradeContainer, 1, 162, 37));
         this.addSlot(new MerchantResultSlot(p_40037_.player, p_40038_, this.tradeContainer, 2, 220, 37));
+        this.player = p_40037_; // CraftBukkit - save player
 
         for (int i = 0; i < 3; i++) {
             for (int j = 0; j < 9; j++) {
@@ -114,12 +_,12 @@
             ItemStack itemstack1 = slot.getItem();
             itemstack = itemstack1.copy();
             if (p_40054_ == 2) {
-                if (!this.moveItemStackTo(itemstack1, 3, 39, true)) {
+                if (!this.moveItemStackTo(itemstack1, 3, 39, true, true)) { // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent
                     return ItemStack.EMPTY;
                 }
 
-                slot.onQuickCraft(itemstack1, itemstack);
-                this.playTradeSound();
+                //  slot.onQuickCraft(itemstack1, itemstack); // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent; moved to after the non-check moveItemStackTo call
+                //  this.playTradeSound();
             } else if (p_40054_ != 0 && p_40054_ != 1) {
                 if (p_40054_ >= 3 && p_40054_ < 30) {
                     if (!this.moveItemStackTo(itemstack1, 30, 39, false)) {
@@ -131,7 +_,7 @@
             } else if (!this.moveItemStackTo(itemstack1, 3, 39, false)) {
                 return ItemStack.EMPTY;
             }
-
+            if (p_40054_ != 2) { // Paper - Add PlayerTradeEvent and PlayerPurchaseEvent; moved down for slot 2
             if (itemstack1.isEmpty()) {
                 slot.setByPlayer(ItemStack.EMPTY);
             } else {
@@ -143,13 +_,28 @@
             }
 
             slot.onTake(p_40053_, itemstack1);
+            } // Paper start - Add PlayerTradeEvent and PlayerPurchaseEvent; handle slot 2
+            if (p_40054_ == 2) { // is merchant result slot
+                slot.onTake(p_40053_, itemstack1);
+                if (itemstack1.isEmpty()) {
+                    slot.set(ItemStack.EMPTY);
+                    return ItemStack.EMPTY;
+                }
+
+                this.moveItemStackTo(itemstack1, 3, 39, true, false); // This should always succeed because it's checked above
+
+                slot.onQuickCraft(itemstack1, itemstack);
+                this.playTradeSound();
+                slot.set(ItemStack.EMPTY); // itemstack1 should ALWAYS be empty
+            }
+            // Paper end - Add PlayerTradeEvent and PlayerPurchaseEvent
         }
 
         return itemstack;
     }
 
     private void playTradeSound() {
-        if (!this.trader.isClientSide()) {
+        if (!this.trader.isClientSide() && this.trader instanceof Entity) { // CraftBukkit - SPIGOT-5035
             Entity entity = (Entity)this.trader;
             entity.level()
                 .playLocalSound(entity.getX(), entity.getY(), entity.getZ(), this.trader.getNotifyTradeSound(), SoundSource.NEUTRAL, 1.0F, 1.0F, false);
