--- a/net/minecraft/world/inventory/ResultContainer.java
+++ b/net/minecraft/world/inventory/ResultContainer.java
@@ -1,5 +_,9 @@
 package net.minecraft.world.inventory;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nullable;
 import net.minecraft.core.NonNullList;
 import net.minecraft.world.Container;
@@ -7,12 +_,50 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.crafting.RecipeHolder;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
 
 public class ResultContainer implements Container, RecipeCraftingHolder {
     private final NonNullList<ItemStack> itemStacks = NonNullList.withSize(1, ItemStack.EMPTY);
     @Nullable
     private RecipeHolder<?> recipeUsed;
 
+    // CraftBukkit start
+    private int maxStack = MAX_STACK;
+
+    public List<ItemStack> getContents() {
+        return this.itemStacks;
+    }
+
+    public InventoryHolder getOwner() {
+        return null; // Result slots don't get an owner
+    }
+
+    // Don't need a transaction; the InventoryCrafting keeps track of it for us
+    public void onOpen(CraftHumanEntity who) {}
+    public void onClose(CraftHumanEntity who) {}
+    public List<HumanEntity> getViewers() {
+        return new ArrayList<>();
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return Math.max(Container.super.getMaxStackSize(), maxStack);
+    }
+
+    @Override
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+
+    @Override
+    public Location getLocation() {
+        return null;
+    }
+    // CraftBukkit end
+
     @Override
     public int getContainerSize() {
         return 1;
@@ -44,8 +_,24 @@
         return ContainerHelper.takeItem(this.itemStacks, 0);
     }
 
+
+    // Youer start
+    AtomicBoolean fakeSetItem = new AtomicBoolean(false);
+    AtomicReference<ItemStack> setItem$itemStack = new AtomicReference<>(null);
+    public void fakeSetItem() {
+        fakeSetItem.set(true);
+    }
+    @Nullable
+    public ItemStack setItem$itemStack() {
+        return setItem$itemStack.getAndSet(null);
+    }
     @Override
     public void setItem(int p_40152_, ItemStack p_40153_) {
+        if (fakeSetItem.getAndSet(false)){
+            setItem$itemStack.set(p_40153_);
+            return;
+        }
+        // Youer end
         this.itemStacks.set(0, p_40153_);
     }
 
