--- a/net/minecraft/world/inventory/StonecutterMenu.java
+++ b/net/minecraft/world/inventory/StonecutterMenu.java
@@ -2,7 +_,6 @@
 
 import com.google.common.collect.Lists;
 import java.util.List;
-import net.minecraft.core.BlockPos;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.world.Container;
@@ -17,6 +_,9 @@
 import net.minecraft.world.item.crafting.StonecutterRecipe;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftInventoryStonecutter;
+import org.bukkit.craftbukkit.inventory.view.CraftStonecutterView;
 
 public class StonecutterMenu extends AbstractContainerMenu {
     public static final int INPUT_SLOT = 0;
@@ -26,7 +_,7 @@
     private static final int USE_ROW_SLOT_START = 29;
     private static final int USE_ROW_SLOT_END = 38;
     private final ContainerLevelAccess access;
-    private final DataSlot selectedRecipeIndex = DataSlot.standalone();
+    private final DataSlot selectedRecipeIndex = DataSlot.shared(new int[1], 0); // Paper - Add PlayerStonecutterRecipeSelectEvent
     private final Level level;
     private List<RecipeHolder<StonecutterRecipe>> recipes = Lists.newArrayList();
     private ItemStack input = ItemStack.EMPTY;
@@ -42,9 +_,32 @@
             StonecutterMenu.this.slotsChanged(this);
             StonecutterMenu.this.slotUpdateListener.run();
         }
+
+        // CraftBukkit start
+        @Override
+        public Location getLocation() {
+            return access.getLocation();
+        }
+        // CraftBukkit end
     };
     final ResultContainer resultContainer = new ResultContainer();
 
+    // CraftBukkit start
+    private CraftStonecutterView bukkitEntity = null;
+    private org.bukkit.entity.Player player;
+
+    @Override
+    public CraftStonecutterView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryStonecutter inventory = new CraftInventoryStonecutter(this.container, this.resultContainer);
+        bukkitEntity = new CraftStonecutterView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
+
     public StonecutterMenu(int p_40294_, Inventory p_40295_) {
         this(p_40294_, p_40295_, ContainerLevelAccess.NULL);
     }
@@ -95,6 +_,7 @@
         }
 
         this.addDataSlot(this.selectedRecipeIndex);
+        player = (org.bukkit.entity.Player) p_40298_.player.getBukkitEntity(); // CraftBukkit
     }
 
     public int getSelectedRecipeIndex() {
@@ -115,13 +_,37 @@
 
     @Override
     public boolean stillValid(Player p_40307_) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, p_40307_, Blocks.STONECUTTER);
     }
 
     @Override
     public boolean clickMenuButton(Player p_40309_, int p_40310_) {
         if (this.isValidRecipeIndex(p_40310_)) {
-            this.selectedRecipeIndex.set(p_40310_);
+            // Paper start - Add PlayerStonecutterRecipeSelectEvent
+            int recipeIndex = p_40310_;
+            this.selectedRecipeIndex.set(recipeIndex);
+            this.selectedRecipeIndex.checkAndClearUpdateFlag(); // mark as changed
+            if (this.isValidRecipeIndex(p_40310_)) {
+                io.papermc.paper.event.player.PlayerStonecutterRecipeSelectEvent event = new io.papermc.paper.event.player.PlayerStonecutterRecipeSelectEvent((org.bukkit.entity.Player) p_40309_.getBukkitEntity(), (org.bukkit.inventory.StonecutterInventory) getBukkitView().getTopInventory(), (org.bukkit.inventory.StonecuttingRecipe) this.getRecipes().get(p_40310_).toBukkitRecipe());
+                if (!event.callEvent()) {
+                    p_40309_.containerMenu.sendAllDataToRemote();
+                    return false;
+                }
+
+                net.minecraft.resources.ResourceLocation key = org.bukkit.craftbukkit.util.CraftNamespacedKey.toMinecraft(event.getStonecuttingRecipe().getKey());
+                if (!this.getRecipes().get(recipeIndex).id().equals(key)) { // If the recipe did NOT stay the same
+                    for (int newRecipeIndex = 0; newRecipeIndex < this.getRecipes().size(); newRecipeIndex++) {
+                        if (this.getRecipes().get(newRecipeIndex).id().equals(key)) {
+                            recipeIndex = newRecipeIndex;
+                            break;
+                        }
+                    }
+                }
+            }
+            p_40309_.containerMenu.sendAllDataToRemote();
+            this.selectedRecipeIndex.set(recipeIndex); // set new index, so that listeners can read it
+            // Paper end - Add PlayerStonecutterRecipeSelectEvent
             this.setupResultSlot();
         }
 
@@ -139,6 +_,7 @@
             this.input = itemstack.copy();
             this.setupRecipeList(p_40302_, itemstack);
         }
+        org.bukkit.craftbukkit.event.CraftEventFactory.callPrepareResultEvent(this, RESULT_SLOT); // Paper - Add PrepareResultEvent
     }
 
     private static SingleRecipeInput createRecipeInput(Container p_346312_) {
