--- a/net/minecraft/world/inventory/StonecutterMenu.java
+++ b/net/minecraft/world/inventory/StonecutterMenu.java
@@ -17,6 +_,9 @@
 import net.minecraft.world.item.crafting.StonecutterRecipe;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.inventory.CraftInventoryStonecutter;
+import org.bukkit.craftbukkit.inventory.view.CraftStonecutterView;
 
 public class StonecutterMenu extends AbstractContainerMenu {
     public static final int INPUT_SLOT = 0;
@@ -45,8 +_,30 @@
             StonecutterMenu.this.slotsChanged(this);
             StonecutterMenu.this.slotUpdateListener.run();
         }
+
+        // CraftBukkit start
+        @Override
+        public Location getLocation() {
+            return access.getLocation();
+        }
+        // CraftBukkit end
     };
     private final ResultContainer resultContainer = new ResultContainer();
+    // CraftBukkit start
+    private CraftStonecutterView bukkitEntity = null;
+    private org.bukkit.entity.Player player;
+
+    @Override
+    public CraftStonecutterView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryStonecutter inventory = new CraftInventoryStonecutter(this.container, this.resultContainer);
+        bukkitEntity = new CraftStonecutterView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
 
     public StonecutterMenu(int containerId, Inventory inventory) {
         this(containerId, inventory, ContainerLevelAccess.NULL);
@@ -92,6 +_,7 @@
         });
         this.addStandardInventorySlots(inventory, 8, 84);
         this.addDataSlot(this.selectedRecipeIndex);
+        player = (org.bukkit.entity.Player) inventory.player.getBukkitEntity(); // CraftBukkit
     }
 
     public int getSelectedRecipeIndex() {
@@ -112,6 +_,7 @@
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return stillValid(this.access, player, Blocks.STONECUTTER);
     }
 
