--- a/net/minecraft/world/item/armortrim/ArmorTrim.java
+++ b/net/minecraft/world/item/armortrim/ArmorTrim.java
@@ -44,7 +_,7 @@
         .withStyle(ChatFormatting.GRAY);
     private final Holder<TrimMaterial> material;
     private final Holder<TrimPattern> pattern;
-    public final boolean showInTooltip;
+    private final boolean showInTooltip;
     private final Function<Holder<ArmorMaterial>, ResourceLocation> innerTexture;
     private final Function<Holder<ArmorMaterial>, ResourceLocation> outerTexture;
 
