--- a/net/minecraft/world/level/pathfinder/Target.java
+++ b/net/minecraft/world/level/pathfinder/Target.java
@@ -1,10 +_,8 @@
 package net.minecraft.world.level.pathfinder;
 
-import io.netty.buffer.ByteBuf;
-import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.network.FriendlyByteBuf;
 
 public class Target extends Node {
-    public static final StreamCodec<ByteBuf, Target> DEBUG_STREAM_CODEC = createDebugStreamCodec(Target::new);
     private float bestHeuristic = Float.MAX_VALUE;
     private Node bestNode;
     private boolean reached;
@@ -34,5 +_,11 @@
 
     public boolean isReached() {
         return this.reached;
+    }
+
+    public static Target createFromStream(FriendlyByteBuf buffer) {
+        Target node = new Target(buffer.readInt(), buffer.readInt(), buffer.readInt());
+        readContents(buffer, node);
+        return node;
     }
 }
