--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -44,6 +_,7 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.util.Vector;
 import org.jspecify.annotations.Nullable;
 
 public abstract class AbstractMinecart extends VehicleEntity {
@@ -97,6 +_,18 @@
         )
     );
 
+    // CraftBukkit start
+    public boolean slowWhenEmpty = true;
+    private double derailedX = 0.5;
+    private double derailedY = 0.5;
+    private double derailedZ = 0.5;
+    private double flyingX = 0.95;
+    private double flyingY = 0.95;
+    private double flyingZ = 0.95;
+    public Double maxSpeed;
+    public double powRailAccelMult = 1.0;
+    // CraftBukkit end
+
     protected AbstractMinecart(EntityType<?> type, Level level) {
         super(type, level);
         this.blocksBuilding = true;
@@ -423,8 +_,8 @@
 
     public Vec3 getRedstoneDirection(BlockPos pos) {
         BlockState state = this.level().getBlockState(pos);
-        if (state.is(Blocks.POWERED_RAIL) && state.getValue(PoweredRailBlock.POWERED)) {
-            RailShape shape = state.getValue(((BaseRailBlock)state.getBlock()).getShapeProperty());
+        if (state.getBlock() instanceof PoweredRailBlock poweredRail && !poweredRail.isActivatorRail() && state.getValue(PoweredRailBlock.POWERED)) {
+            RailShape shape = ((BaseRailBlock)state.getBlock()).getRailDirection(state, this.level(), pos, this);
             if (shape == RailShape.EAST_WEST) {
                 if (this.isRedstoneConductor(pos.west())) {
                     return new Vec3(1.0, 0.0, 0.0);
@@ -596,4 +_,26 @@
     public boolean isFurnace() {
         return false;
     }
+
+    // CraftBukkit start - Methods for getting and setting flying and derailed velocity modifiers
+    public Vector getFlyingVelocityMod() {
+        return new Vector(flyingX, flyingY, flyingZ);
+    }
+
+    public void setFlyingVelocityMod(Vector flying) {
+        flyingX = flying.getX();
+        flyingY = flying.getY();
+        flyingZ = flying.getZ();
+    }
+
+    public Vector getDerailedVelocityMod() {
+        return new Vector(derailedX, derailedY, derailedZ);
+    }
+
+    public void setDerailedVelocityMod(Vector derailed) {
+        derailedX = derailed.getX();
+        derailedY = derailed.getY();
+        derailedZ = derailed.getZ();
+    }
+    // CraftBukkit end
 }
