--- a/net/minecraft/world/entity/vehicle/minecart/MinecartTNT.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartTNT.java
@@ -24,6 +_,8 @@
 import net.minecraft.world.level.material.FluidState;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 import org.jspecify.annotations.Nullable;
 
 public class MinecartTNT extends AbstractMinecart {
@@ -35,6 +_,7 @@
     public int fuse = -1;
     public float explosionPowerBase = 4.0F;
     public float explosionSpeedFactor = 1.0F;
+    public boolean isIncendiary = false; // CraftBukkit - add field
 
     public MinecartTNT(EntityType<? extends MinecartTNT> type, Level level) {
         super(type, level);
@@ -100,20 +_,18 @@
         if (this.level() instanceof ServerLevel level) {
             if (level.getGameRules().get(GameRules.TNT_EXPLODES)) {
                 double speed = Math.min(Math.sqrt(speedSqr), 5.0);
-                level.explode(
-                    this,
-                    damageSource,
-                    null,
-                    this.getX(),
-                    this.getY(),
-                    this.getZ(),
-                    (float)(this.explosionPowerBase + this.explosionSpeedFactor * this.random.nextDouble() * 1.5 * speed),
-                    false,
-                    Level.ExplosionInteraction.TNT
-                );
-                this.discard();
+                // CraftBukkit start
+                ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), (float)(this.explosionPowerBase + this.explosionSpeedFactor * this.random.nextDouble() * 1.5 * speed), this.isIncendiary);
+                level.getCraftServer().getPluginManager().callEvent(event);
+                if (event.isCancelled()) {
+                    fuse = -1;
+                    return;
+                }
+                level.explode(this, damageSource, null, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.TNT);
+                // CraftBukkit end
+                this.discard(EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit - add Bukkit remove cause
             } else if (this.isPrimed()) {
-                this.discard();
+                this.discard(EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit - add Bukkit remove cause
             }
         }
     }
