--- a/net/minecraft/world/entity/vehicle/minecart/MinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/MinecartBehavior.java
@@ -7,12 +_,17 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.properties.RailShape;
 import net.minecraft.world.phys.Vec3;
+import org.jspecify.annotations.Nullable;
 
 public abstract class MinecartBehavior {
     protected final AbstractMinecart minecart;
 
     protected MinecartBehavior(AbstractMinecart minecart) {
         this.minecart = minecart;
+    }
+
+    public @Nullable InterpolationHandler getInterpolation() {
+        return null;
     }
 
     public void onInterpolationStart(InterpolationHandler interpolation) {
