--- a/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/NewMinecartBehavior.java
@@ -26,6 +_,11 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.entity.Vehicle;
+import org.bukkit.event.vehicle.VehicleEntityCollisionEvent;
+// CraftBukkit end
+
 public class NewMinecartBehavior extends MinecartBehavior {
     public static final int POS_ROT_LERP_TICKS = 3;
     public static final double ON_RAIL_Y_OFFSET = 0.1;
@@ -153,7 +_,7 @@
 
     public void adjustToRails(BlockPos targetBlockPos, BlockState currentState, boolean instant) {
         if (BaseRailBlock.isRail(currentState)) {
-            RailShape shape = currentState.getValue(((BaseRailBlock)currentState.getBlock()).getShapeProperty());
+            RailShape shape = ((BaseRailBlock)currentState.getBlock()).getRailDirection(currentState, this.level(), targetBlockPos, this.minecart);
             Pair<Vec3i, Vec3i> exits = AbstractMinecart.exits(shape);
             Vec3 exit0 = new Vec3(exits.getFirst()).scale(0.5);
             Vec3 exit1 = new Vec3(exits.getSecond()).scale(0.5);
@@ -246,12 +_,12 @@
             if (onRails) {
                 this.minecart.resetFallDistance();
                 this.minecart.setOldPosAndRot();
-                if (currentState.is(Blocks.ACTIVATOR_RAIL)) {
+                if (currentState.getBlock() instanceof PoweredRailBlock poweredRail && poweredRail.isActivatorRail()) {
                     this.minecart
                         .activateMinecart(level, currentPos.getX(), currentPos.getY(), currentPos.getZ(), currentState.getValue(PoweredRailBlock.POWERED));
                 }
 
-                RailShape shape = currentState.getValue(((BaseRailBlock)currentState.getBlock()).getShapeProperty());
+                RailShape shape = ((BaseRailBlock)currentState.getBlock()).getRailDirection(currentState, this.level(), currentPos, this.minecart);;
                 Vec3 newDeltaMovement = this.calculateTrackSpeed(level, initialStepDeltaMovement.horizontal(), trackIteration, currentPos, currentState, shape);
                 if (trackIteration.firstIteration) {
                     trackIteration.movementLeft = newDeltaMovement.horizontalDistance();
@@ -381,7 +_,7 @@
     }
 
     private Vec3 calculateHaltTrackSpeed(Vec3 deltaMovement, BlockState state) {
-        if (state.is(Blocks.POWERED_RAIL) && !state.getValue(PoweredRailBlock.POWERED)) {
+        if (state.getBlock() instanceof PoweredRailBlock poweredRail && !poweredRail.isActivatorRail() && !state.getValue(PoweredRailBlock.POWERED)) {
             return deltaMovement.length() < 0.03 ? Vec3.ZERO : deltaMovement.scale(0.5);
         } else {
             return deltaMovement;
@@ -389,7 +_,7 @@
     }
 
     private Vec3 calculateBoostTrackSpeed(Vec3 deltaMovement, BlockPos pos, BlockState state) {
-        if (state.is(Blocks.POWERED_RAIL) && state.getValue(PoweredRailBlock.POWERED)) {
+        if (state.getBlock() instanceof PoweredRailBlock poweredRail && !poweredRail.isActivatorRail() && state.getValue(PoweredRailBlock.POWERED)) {
             if (deltaMovement.length() > 0.01) {
                 return deltaMovement.normalize().scale(deltaMovement.length() + 0.06);
             }
@@ -440,10 +_,11 @@
         }
 
         this.minecart.move(MoverType.SELF, newPosition.subtract(oldPosition));
-        BlockState newBlockState = this.level().getBlockState(BlockPos.containing(newPosition));
+        BlockPos railPos = BlockPos.containing(newPosition);
+        BlockState newBlockState = this.level().getBlockState(railPos);
         if (inHill) {
             if (BaseRailBlock.isRail(newBlockState)) {
-                RailShape newRailShape = newBlockState.getValue(((BaseRailBlock)newBlockState.getBlock()).getShapeProperty());
+                RailShape newRailShape = ((BaseRailBlock)newBlockState.getBlock()).getRailDirection(newBlockState, this.level(), railPos, this.minecart);
                 if (this.restAtVShape(shape, newRailShape)) {
                     return 0.0;
                 }
@@ -479,6 +_,12 @@
 
     @Override
     public double getMaxSpeed(ServerLevel level) {
+        // CraftBukkit start
+        Double maxSpeed = this.minecart.maxSpeed;
+        if (maxSpeed != null) {
+            return (this.minecart.isInWater() ? maxSpeed / 2.0D : maxSpeed);
+        }
+        // CraftBukkit end
         return level.getGameRules().get(GameRules.MAX_MINECART_SPEED).intValue() * (this.minecart.isInWater() ? 0.5 : 1.0) / 20.0;
     }
 
@@ -494,7 +_,7 @@
 
     @Override
     public double getSlowdownFactor() {
-        return this.minecart.isVehicle() ? 0.997 : 0.975;
+        return this.minecart.isVehicle() || !this.minecart.slowWhenEmpty ? 0.997D : 0.975D; // CraftBukkit - add !this.slowWhenEmpty
     }
 
     @Override
@@ -518,6 +_,14 @@
                         && !(entity instanceof AbstractMinecart)
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
+                        // CraftBukkit start
+                        VehicleEntityCollisionEvent collisionEvent = new VehicleEntityCollisionEvent((Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity());
+                        this.level().getCraftServer().getPluginManager().callEvent(collisionEvent);
+
+                        if (collisionEvent.isCancelled()) {
+                            continue;
+                        }
+                        // CraftBukkit end
                         boolean pickedUp = entity.startRiding(this.minecart);
                         if (pickedUp) {
                             return true;
@@ -541,15 +_,33 @@
                         || entity instanceof AbstractMinecart
                         || this.minecart.isVehicle()
                         || entity.isPassenger()) {
+                        // CraftBukkit start
+                        if (!this.minecart.isPassengerOfSameVehicle(entity)) {
+                            VehicleEntityCollisionEvent collisionEvent = new VehicleEntityCollisionEvent((Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity());
+                            this.level().getCraftServer().getPluginManager().callEvent(collisionEvent);
+
+                            if (collisionEvent.isCancelled()) {
+                                continue;
+                            }
+                        }
+                        // CraftBukkit end
                         entity.push(this.minecart);
                         pushed = true;
                     }
                 }
             }
         } else {
-            for (Entity entity : this.level().getEntities(this.minecart, hitbox)) {
-                if (!this.minecart.hasPassenger(entity) && entity.isPushable() && entity instanceof AbstractMinecart) {
-                    entity.push(this.minecart);
+            for (Entity entityx : this.level().getEntities(this.minecart, hitbox)) {
+                if (!this.minecart.hasPassenger(entityx) && entityx.isPushable() && entityx instanceof AbstractMinecart) {
+                    // CraftBukkit start
+                    VehicleEntityCollisionEvent collisionEvent = new VehicleEntityCollisionEvent((Vehicle) this.minecart.getBukkitEntity(), entityx.getBukkitEntity());
+                    this.level().getCraftServer().getPluginManager().callEvent(collisionEvent);
+
+                    if (collisionEvent.isCancelled()) {
+                        continue;
+                    }
+                    // CraftBukkit end
+                    entityx.push(this.minecart);
                     pushed = true;
                 }
             }
