--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -24,6 +_,11 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.entity.Vehicle;
+import org.bukkit.event.vehicle.VehicleEntityCollisionEvent;
+// CraftBukkit end
+
 public class OldMinecartBehavior extends MinecartBehavior {
     private static final double MINECART_RIDABLE_THRESHOLD = 0.01;
     private static final double MAX_SPEED_IN_WATER = 0.2;
@@ -62,7 +_,7 @@
             this.minecart.setOnRails(onRails);
             if (onRails) {
                 this.moveAlongTrack(level);
-                if (state.is(Blocks.ACTIVATOR_RAIL)) {
+                if (state.getBlock() instanceof PoweredRailBlock poweredRail && poweredRail.isActivatorRail()) {
                     this.minecart.activateMinecart(level, var11.getX(), var11.getY(), var11.getZ(), state.getValue(PoweredRailBlock.POWERED));
                 }
             } else {
@@ -112,7 +_,7 @@
         y = pos.getY();
         boolean powerTrack = false;
         boolean haltTrack = false;
-        if (state.is(Blocks.POWERED_RAIL)) {
+        if (state.getBlock() instanceof PoweredRailBlock poweredRail && !poweredRail.isActivatorRail()) {
             powerTrack = state.getValue(PoweredRailBlock.POWERED);
             haltTrack = !powerTrack;
         }
@@ -123,7 +_,7 @@
         }
 
         Vec3 movement = this.getDeltaMovement();
-        RailShape shape = state.getValue(((BaseRailBlock)state.getBlock()).getShapeProperty());
+        RailShape shape = ((BaseRailBlock)state.getBlock()).getRailDirection(state, this.level(), pos, this.minecart);;
         switch (shape) {
             case ASCENDING_EAST:
                 this.setDeltaMovement(movement.add(-slideSpeed, 0.0, 0.0));
@@ -279,9 +_,10 @@
             yt--;
         }
 
-        BlockState state = this.level().getBlockState(new BlockPos(xt, yt, zt));
+        var railPos = new BlockPos(xt, yt, zt);
+        BlockState state = this.level().getBlockState(railPos);
         if (BaseRailBlock.isRail(state)) {
-            RailShape shape = state.getValue(((BaseRailBlock)state.getBlock()).getShapeProperty());
+            RailShape shape = ((BaseRailBlock)state.getBlock()).getRailDirection(state, this.level(), railPos, this.minecart);
             y = yt;
             if (shape.isSlope()) {
                 y = yt + 1;
@@ -317,9 +_,10 @@
             yt--;
         }
 
-        BlockState state = this.level().getBlockState(new BlockPos(xt, yt, zt));
+        BlockPos railPos = new BlockPos(xt, yt, zt);
+        BlockState state = this.level().getBlockState(railPos);
         if (BaseRailBlock.isRail(state)) {
-            RailShape shape = state.getValue(((BaseRailBlock)state.getBlock()).getShapeProperty());
+            RailShape shape = ((BaseRailBlock)state.getBlock()).getRailDirection(state, this.level(), railPos, this.minecart);
             Pair<Vec3i, Vec3i> exits = AbstractMinecart.exits(shape);
             Vec3i exit0 = exits.getFirst();
             Vec3i exit1 = exits.getSecond();
@@ -375,16 +_,42 @@
                         && !(entity instanceof AbstractMinecart)
                         && !this.minecart.isVehicle()
                         && !entity.isPassenger()) {
+                        // CraftBukkit start
+                        VehicleEntityCollisionEvent collisionEvent = new VehicleEntityCollisionEvent((Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity());
+                        this.level().getCraftServer().getPluginManager().callEvent(collisionEvent);
+
+                        if (collisionEvent.isCancelled()) {
+                            continue;
+                        }
+                        // CraftBukkit end
                         entity.startRiding(this.minecart);
                     } else {
+                        // CraftBukkit start
+                        if (!this.minecart.isPassengerOfSameVehicle(entity)) {
+                            VehicleEntityCollisionEvent collisionEvent = new VehicleEntityCollisionEvent((Vehicle) this.minecart.getBukkitEntity(), entity.getBukkitEntity());
+                            this.level().getCraftServer().getPluginManager().callEvent(collisionEvent);
+
+                            if (collisionEvent.isCancelled()) {
+                                continue;
+                            }
+                        }
+                        // CraftBukkit end
                         entity.push(this.minecart);
                     }
                 }
             }
         } else {
-            for (Entity entity : this.level().getEntities(this.minecart, hitbox)) {
-                if (!this.minecart.hasPassenger(entity) && entity.isPushable() && entity instanceof AbstractMinecart) {
-                    entity.push(this.minecart);
+            for (Entity entityx : this.level().getEntities(this.minecart, hitbox)) {
+                if (!this.minecart.hasPassenger(entityx) && entityx.isPushable() && entityx instanceof AbstractMinecart) {
+                    // CraftBukkit start
+                    VehicleEntityCollisionEvent collisionEvent = new VehicleEntityCollisionEvent((Vehicle) this.minecart.getBukkitEntity(), entityx.getBukkitEntity());
+                    this.level().getCraftServer().getPluginManager().callEvent(collisionEvent);
+
+                    if (collisionEvent.isCancelled()) {
+                        continue;
+                    }
+                    // CraftBukkit end
+                    entityx.push(this.minecart);
                 }
             }
         }
@@ -406,11 +_,17 @@
 
     @Override
     public double getMaxSpeed(ServerLevel level) {
+        // CraftBukkit start
+        Double maxSpeed = this.minecart.maxSpeed;
+        if (maxSpeed != null) {
+            return (this.minecart.isInWater() ? maxSpeed / 2.0D : maxSpeed);
+        }
+        // CraftBukkit end
         return this.minecart.isInWater() ? 0.2 : 0.4;
     }
 
     @Override
     public double getSlowdownFactor() {
-        return this.minecart.isVehicle() ? 0.997 : 0.96;
+        return this.minecart.isVehicle() || !this.minecart.slowWhenEmpty ? 0.997D : 0.96D; // CraftBukkit - add !this.slowWhenEmpty
     }
 }
