--- a/net/minecraft/client/gui/font/FontManager.java
+++ b/net/minecraft/client/gui/font/FontManager.java
@@ -231,9 +_,6 @@
         for (FontSet value : this.fontSets.values()) {
             value.reload(fontOptions);
         }
-
-        this.anyGlyphs.invalidate();
-        this.nonFishyGlyphs.invalidate();
     }
 
     private static List<Pair<FontManager.BuilderId, GlyphProviderDefinition.Conditional>> loadResourceStack(List<Resource> resourceStack, Identifier fontName) {
