--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -20,6 +20,16 @@
     public static final TicketType PORTAL = register("portal", 300L, 15);
     public static final TicketType ENDER_PEARL = register("ender_pearl", 40L, 14);
     public static final TicketType UNKNOWN = register("unknown", 1L, 18);
+    // CraftBukkit start
+    public static long pluginTimeout = 0L;
+    public static final TicketType PLUGIN = register("plugin", 0L, 14);
+    public static final TicketType PLUGIN_TICKET = register("plugin_ticket", 0L, 14);
+
+    @Override
+    public long timeout() {
+        return (this != TicketType.PLUGIN) ? this.timeout : TicketType.pluginTimeout;
+    }
+    // CraftBukkit end
 
     private static TicketType register(String s, long i, int j) {
         return (TicketType) IRegistry.register(BuiltInRegistries.TICKET_TYPE, s, new TicketType(i, j));
