--- a/net/minecraft/world/entity/animal/frog/FrogAi.java
+++ b/net/minecraft/world/entity/animal/frog/FrogAi.java
@@ -9,7 +_,6 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.tags.BlockTags;
-import net.minecraft.tags.FluidTags;
 import net.minecraft.tags.ItemTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.valueproviders.UniformInt;
@@ -36,7 +_,7 @@
 import net.minecraft.world.entity.ai.behavior.StartAttacking;
 import net.minecraft.world.entity.ai.behavior.StopAttackingIfTargetInvalid;
 import net.minecraft.world.entity.ai.behavior.TryFindLand;
-import net.minecraft.world.entity.ai.behavior.TryFindLandNearLiquid;
+import net.minecraft.world.entity.ai.behavior.TryFindLandNearWater;
 import net.minecraft.world.entity.ai.behavior.TryLaySpawnOnFluidNearLand;
 import net.minecraft.world.entity.ai.behavior.declarative.BehaviorBuilder;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
@@ -153,7 +_,7 @@
                 Pair.of(
                     1, StartAttacking.create((level, body) -> canAttack(body), (level, body) -> body.getBrain().getMemory(MemoryModuleType.NEAREST_ATTACKABLE))
                 ),
-                Pair.of(2, TryFindLandNearLiquid.create(8, 1.0F, FluidTags.FROG_TRIES_TO_FIND_LAND_NEAR)),
+                Pair.of(2, TryFindLandNearWater.create(8, 1.0F)),
                 Pair.of(3, TryLaySpawnOnFluidNearLand.create(Blocks.FROGSPAWN)),
                 Pair.of(
                     4,
