--- a/net/minecraft/world/entity/animal/frog/FrogVariant.java
+++ b/net/minecraft/world/entity/animal/frog/FrogVariant.java
@@ -6,10 +_,10 @@
 import net.minecraft.core.ClientAsset;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.world.entity.variant.PriorityProvider;
 import net.minecraft.world.entity.variant.SpawnCondition;
 import net.minecraft.world.entity.variant.SpawnContext;
@@ -27,7 +_,7 @@
     public static final Codec<FrogVariant> NETWORK_CODEC = RecordCodecBuilder.create(
         i -> i.group(ClientAsset.ResourceTexture.DEFAULT_FIELD_CODEC.forGetter(FrogVariant::assetInfo)).apply(i, FrogVariant::new)
     );
-    public static final Codec<Holder<FrogVariant>> CODEC = RegistryCodecs.holder(Registries.FROG_VARIANT);
+    public static final Codec<Holder<FrogVariant>> CODEC = RegistryFixedCodec.create(Registries.FROG_VARIANT);
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<FrogVariant>> STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.FROG_VARIANT);
 
     private FrogVariant(ClientAsset.ResourceTexture assetInfo) {
