--- a/net/minecraft/world/entity/animal/frog/ShootTongue.java
+++ b/net/minecraft/world/entity/animal/frog/ShootTongue.java
@@ -92,7 +_,8 @@
             if (entity.isAlive()) {
                 p_218642_.doHurtTarget(entity);
                 if (!entity.isAlive()) {
-                    entity.remove(Entity.RemovalReason.KILLED);
+                    entity.setRemoved$cause(org.bukkit.event.entity.EntityRemoveEvent.Cause.DEATH);
+                    entity.remove(Entity.RemovalReason.KILLED); // CraftBukkit - add Bukkit remove cause
                 }
             }
         }
