--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -134,19 +_,19 @@
 
     @Nullable
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return null;
     }
 
     @Nullable
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_218713_) {
+    public SoundEvent getHurtSound(DamageSource p_218713_) {
         return SoundEvents.TADPOLE_HURT;
     }
 
     @Nullable
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.TADPOLE_DEATH;
     }
 
@@ -222,8 +_,10 @@
     }
 
     private void ageUp() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.FROG, (timer) -> {})) return;
         if (this.level() instanceof ServerLevel serverlevel) {
             this.convertTo(EntityType.FROG, ConversionParams.single(this, false, false), p_432547_ -> {
+                net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, p_432547_);
                 p_432547_.finalizeSpawn(serverlevel, this.level().getCurrentDifficultyAt(p_432547_.blockPosition()), EntitySpawnReason.CONVERSION, null);
                 p_432547_.setPersistenceRequired();
                 p_432547_.fudgePositionAfterSizeChange(this.getDimensions(this.getPose()));
