--- a/net/minecraft/client/gui/screens/multiplayer/JoinMultiplayerScreen.java
+++ b/net/minecraft/client/gui/screens/multiplayer/JoinMultiplayerScreen.java
@@ -193,7 +_,7 @@
             ServerData current = onlineServerEntry.getServerData();
             current.name = this.editingServer.name;
             current.ip = this.editingServer.ip;
-            current.setState(ServerData.State.INITIAL);
+            current.copyFrom(this.editingServer);
             this.servers.save();
             this.serverSelectionList.updateOnlineServers(this.servers);
         }
