--- a/net/minecraft/world/level/levelgen/structure/templatesystem/loader/TemplateSource.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/loader/TemplateSource.java
@@ -57,7 +_,7 @@
         }
     }
 
-    public static CompoundTag readStructure(InputStream input) throws IOException {
+    private static CompoundTag readStructure(InputStream input) throws IOException {
         return NbtIo.readCompressed(input, NbtAccounter.unlimitedHeap());
     }
 
