--- a/net/minecraft/world/level/block/grower/TreeGrower.java
+++ b/net/minecraft/world/level/block/grower/TreeGrower.java
@@ -17,9 +_,13 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.SaplingBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
+import org.bukkit.TreeType;
+import org.bukkit.craftbukkit.block.CapturedBlockState;
+import org.bukkit.craftbukkit.util.CraftLocation;
 
 public final class TreeGrower {
     private static final Map<String, TreeGrower> GROWERS = new Object2ObjectArrayMap<>();
@@ -103,6 +_,37 @@
         GROWERS.put(p_304522_, this);
     }
 
+    public TreeGrower youer(ServerLevel serverLevel) {
+        serverLevel.captureTreeGeneration = true;
+        serverLevel.captureTreeGeneration$Youer = true;
+        return this;
+    }
+
+    public void clearCaptureTreeGeneration(ServerLevel serverLevel, BlockPos blockPos) {
+        if (serverLevel.captureTreeGeneration$Youer) {
+            serverLevel.captureTreeGeneration = false;
+            serverLevel.captureTreeGeneration$Youer = false;
+            java.util.List<org.bukkit.craftbukkit.block.CraftBlockState> capturedBlockStates = new java.util.ArrayList<>(serverLevel.capturedBlockStates.values());
+            serverLevel.capturedBlockStates.clear();
+            TreeType treeType = SaplingBlock.treeType;
+            SaplingBlock.treeType = null;
+
+            if (!capturedBlockStates.isEmpty()) {
+                org.bukkit.Location location = CraftLocation.toBukkit(blockPos, serverLevel.getWorld());
+                org.bukkit.event.world.StructureGrowEvent event = null;
+                if (treeType != null) {
+                    event = new org.bukkit.event.world.StructureGrowEvent(location, treeType, false, null, (java.util.List<org.bukkit.block.BlockState>) (java.util.List<? extends org.bukkit.block.BlockState>) capturedBlockStates);
+                    org.bukkit.Bukkit.getPluginManager().callEvent(event);
+                }
+                if (event == null || !event.isCancelled()) {
+                    for (org.bukkit.block.BlockState blockstate : capturedBlockStates) {
+                        CapturedBlockState.setBlockState(blockstate);
+                    }
+                }
+            }
+        }
+    }
+
     @Nullable
     private ResourceKey<ConfiguredFeature<?, ?>> getConfiguredFeature(RandomSource p_304525_, boolean p_304410_) {
         if (p_304525_.nextFloat() < this.secondaryChance) {
@@ -130,7 +_,14 @@
                 .registryOrThrow(Registries.CONFIGURED_FEATURE)
                 .getHolder(resourcekey)
                 .orElse(null);
+            var event = net.neoforged.neoforge.event.EventHooks.fireBlockGrowFeature(p_304396_, p_304893_, p_304643_, holder);
+            holder = event.getFeature();
+            if (event.isCanceled()){
+                clearCaptureTreeGeneration(p_304396_, p_304643_);
+                return false;
+            }
             if (holder != null) {
+                setTreeType(holder); // CraftBukkit
                 for (int i = 0; i >= -1; i--) {
                     for (int j = 0; j >= -1; j--) {
                         if (isTwoByTwoSapling(p_304439_, p_304396_, p_304643_, i, j)) {
@@ -141,6 +_,7 @@
                             p_304396_.setBlock(p_304643_.offset(i, 0, j + 1), blockstate, 4);
                             p_304396_.setBlock(p_304643_.offset(i + 1, 0, j + 1), blockstate, 4);
                             if (configuredfeature.place(p_304396_, p_304672_, p_304893_, p_304643_.offset(i, 0, j))) {
+                                clearCaptureTreeGeneration(p_304396_, p_304643_);
                                 return true;
                             }
 
@@ -148,6 +_,7 @@
                             p_304396_.setBlock(p_304643_.offset(i + 1, 0, j), p_304439_, 4);
                             p_304396_.setBlock(p_304643_.offset(i, 0, j + 1), p_304439_, 4);
                             p_304396_.setBlock(p_304643_.offset(i + 1, 0, j + 1), p_304439_, 4);
+                            clearCaptureTreeGeneration(p_304396_, p_304643_);
                             return false;
                         }
                     }
@@ -157,15 +_,24 @@
 
         ResourceKey<ConfiguredFeature<?, ?>> resourcekey1 = this.getConfiguredFeature(p_304893_, this.hasFlowers(p_304396_, p_304643_));
         if (resourcekey1 == null) {
+            clearCaptureTreeGeneration(p_304396_, p_304643_);
             return false;
         } else {
             Holder<ConfiguredFeature<?, ?>> holder1 = p_304396_.registryAccess()
                 .registryOrThrow(Registries.CONFIGURED_FEATURE)
                 .getHolder(resourcekey1)
                 .orElse(null);
+            var event = net.neoforged.neoforge.event.EventHooks.fireBlockGrowFeature(p_304396_, p_304893_, p_304643_, holder1);
+            holder1 = event.getFeature();
+            if (event.isCanceled()) {
+                clearCaptureTreeGeneration(p_304396_, p_304643_);
+                return false;
+            }
             if (holder1 == null) {
+                clearCaptureTreeGeneration(p_304396_, p_304643_);
                 return false;
             } else {
+                setTreeType(holder1); // CraftBukkit
                 ConfiguredFeature<?, ?> configuredfeature1 = holder1.value();
                 BlockState blockstate1 = p_304396_.getFluidState(p_304643_).createLegacyBlock();
                 p_304396_.setBlock(p_304643_, blockstate1, 4);
@@ -173,10 +_,11 @@
                     if (p_304396_.getBlockState(p_304643_) == blockstate1) {
                         p_304396_.sendBlockUpdated(p_304643_, p_304439_, blockstate1, 2);
                     }
-
+                    clearCaptureTreeGeneration(p_304396_, p_304643_);
                     return true;
                 } else {
                     p_304396_.setBlock(p_304643_, p_304439_, 4);
+                    clearCaptureTreeGeneration(p_304396_, p_304643_);
                     return false;
                 }
             }
@@ -200,4 +_,63 @@
 
         return false;
     }
+
+    // CraftBukkit start
+    public static Map<String, TreeGrower> getGrowers() {
+        return GROWERS;
+    }
+
+    private void setTreeType(Holder<ConfiguredFeature<?, ?>> holder) {
+        ResourceKey<ConfiguredFeature<?, ?>> worldgentreeabstract = holder.unwrapKey().get();
+        if (worldgentreeabstract == TreeFeatures.OAK || worldgentreeabstract == TreeFeatures.OAK_BEES_005) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TREE;
+        } else if (worldgentreeabstract == TreeFeatures.HUGE_RED_MUSHROOM) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.RED_MUSHROOM;
+        } else if (worldgentreeabstract == TreeFeatures.HUGE_BROWN_MUSHROOM) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BROWN_MUSHROOM;
+        } else if (worldgentreeabstract == TreeFeatures.JUNGLE_TREE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.COCOA_TREE;
+        } else if (worldgentreeabstract == TreeFeatures.JUNGLE_TREE_NO_VINE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SMALL_JUNGLE;
+        } else if (worldgentreeabstract == TreeFeatures.PINE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_REDWOOD;
+        } else if (worldgentreeabstract == TreeFeatures.SPRUCE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.REDWOOD;
+        } else if (worldgentreeabstract == TreeFeatures.ACACIA) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.ACACIA;
+        } else if (worldgentreeabstract == TreeFeatures.BIRCH || worldgentreeabstract == TreeFeatures.BIRCH_BEES_005) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIRCH;
+        } else if (worldgentreeabstract == TreeFeatures.SUPER_BIRCH_BEES_0002) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_BIRCH;
+        } else if (worldgentreeabstract == TreeFeatures.SWAMP_OAK) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.SWAMP;
+        } else if (worldgentreeabstract == TreeFeatures.FANCY_OAK || worldgentreeabstract == TreeFeatures.FANCY_OAK_BEES_005) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.BIG_TREE;
+        } else if (worldgentreeabstract == TreeFeatures.JUNGLE_BUSH) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE_BUSH;
+        } else if (worldgentreeabstract == TreeFeatures.DARK_OAK) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.DARK_OAK;
+        } else if (worldgentreeabstract == TreeFeatures.MEGA_SPRUCE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_REDWOOD;
+        } else if (worldgentreeabstract == TreeFeatures.MEGA_PINE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MEGA_PINE;
+        } else if (worldgentreeabstract == TreeFeatures.MEGA_JUNGLE_TREE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.JUNGLE;
+        } else if (worldgentreeabstract == TreeFeatures.AZALEA_TREE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.AZALEA;
+        } else if (worldgentreeabstract == TreeFeatures.MANGROVE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.MANGROVE;
+        } else if (worldgentreeabstract == TreeFeatures.TALL_MANGROVE) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.TALL_MANGROVE;
+        } else if (worldgentreeabstract == TreeFeatures.CHERRY || worldgentreeabstract == TreeFeatures.CHERRY_BEES_005) {
+            net.minecraft.world.level.block.SaplingBlock.treeType = org.bukkit.TreeType.CHERRY;
+        } else {
+            org.bukkit.TreeType treeType = com.mohistmc.youer.neoforge.NeoForgeInjectBukkit.treeTypeByGrowerName.get(this.name);
+            if (treeType == null) {
+                treeType = org.bukkit.TreeType.CUSTOM;
+            }
+            net.minecraft.world.level.block.SaplingBlock.treeType = treeType;
+        }
+    }
+    // CraftBukkit end
 }
