--- a/net/minecraft/world/level/block/grower/TreeGrower.java
+++ b/net/minecraft/world/level/block/grower/TreeGrower.java
@@ -23,6 +_,11 @@
 import net.minecraft.world.level.levelgen.feature.configurations.TreeConfiguration;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import net.minecraft.world.level.block.SaplingBlock;
+import org.bukkit.TreeType;
+// CraftBukkit end
+
 public final class TreeGrower {
     private static final Map<String, TreeGrower> GROWERS = new Object2ObjectArrayMap<>();
     public static final Codec<TreeGrower> CODEC = Codec.stringResolver(g -> g.name, GROWERS::get);
@@ -131,7 +_,11 @@
                 .lookupOrThrow(Registries.CONFIGURED_FEATURE)
                 .get(megaFeatureKey)
                 .orElse(null);
+            var event = net.neoforged.neoforge.event.EventHooks.fireBlockGrowFeature(level, random, pos, featureHolder);
+            featureHolder = event.getFeature();
+            if (event.isCanceled()) return false;
             if (featureHolder != null) {
+                setTreeType(featureHolder); // CraftBukkit
                 for (int dx = 0; dx >= -1; dx--) {
                     for (int dz = 0; dz >= -1; dz--) {
                         if (isTwoByTwoSapling(state, level, pos, dx, dz)) {
@@ -159,25 +_,29 @@
         ResourceKey<ConfiguredFeature<?, ?>> featureKey = this.getConfiguredFeature(random, this.hasFlowers(level, pos));
         if (featureKey == null) {
             return false;
-        }
-
-        Holder<ConfiguredFeature<?, ?>> featureHolder = level.registryAccess().lookupOrThrow(Registries.CONFIGURED_FEATURE).get(featureKey).orElse(null);
-        if (featureHolder == null) {
-            return false;
-        }
-
-        ConfiguredFeature<?, ?> feature = featureHolder.value();
-        BlockState emptyBlock = level.getFluidState(pos).createLegacyBlock();
-        level.setBlock(pos, emptyBlock, 260);
-        if (feature.place(level, generator, random, pos)) {
-            if (level.getBlockState(pos) == emptyBlock) {
-                level.sendBlockUpdated(pos, state, emptyBlock, 2);
-            }
-
-            return true;
         } else {
-            level.setBlock(pos, state, 260);
-            return false;
+            Holder<ConfiguredFeature<?, ?>> featureHolder = level.registryAccess().lookupOrThrow(Registries.CONFIGURED_FEATURE).get(featureKey).orElse(null);
+            var event = net.neoforged.neoforge.event.EventHooks.fireBlockGrowFeature(level, random, pos, featureHolder);
+            featureHolder = event.getFeature();
+            if (event.isCanceled()) return false;
+            if (featureHolder == null) {
+                return false;
+            } else {
+                setTreeType(featureHolder); // CraftBukkit
+                ConfiguredFeature<?, ?> feature = featureHolder.value();
+                BlockState emptyBlock = level.getFluidState(pos).createLegacyBlock();
+                level.setBlock(pos, emptyBlock, 260);
+                if (feature.place(level, generator, random, pos)) {
+                    if (level.getBlockState(pos) == emptyBlock) {
+                        level.sendBlockUpdated(pos, state, emptyBlock, 2);
+                    }
+
+                    return true;
+                } else {
+                    level.setBlock(pos, state, 260);
+                    return false;
+                }
+            }
         }
     }
 
@@ -210,4 +_,60 @@
             ? OptionalInt.of(treeConfig.trunkPlacer.getBaseHeight())
             : OptionalInt.empty();
     }
+
+    // CraftBukkit start
+    private void setTreeType(Holder<ConfiguredFeature<?, ?>> feature) {
+        org.bukkit.TreeType treeType; // SparklyPaper - parallel world ticking
+        if (feature.is(TreeFeatures.OAK) || feature.is(TreeFeatures.OAK_BEES_005)) {
+            treeType = org.bukkit.TreeType.TREE; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.HUGE_RED_MUSHROOM)) {
+            treeType = org.bukkit.TreeType.RED_MUSHROOM; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.HUGE_BROWN_MUSHROOM)) {
+            treeType = org.bukkit.TreeType.BROWN_MUSHROOM; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.JUNGLE_TREE)) {
+            treeType = org.bukkit.TreeType.COCOA_TREE; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.JUNGLE_TREE_NO_VINE)) {
+            treeType = org.bukkit.TreeType.SMALL_JUNGLE; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.PINE)) {
+            treeType = org.bukkit.TreeType.TALL_REDWOOD; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.SPRUCE)) {
+            treeType = org.bukkit.TreeType.REDWOOD; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.ACACIA)) {
+            treeType = org.bukkit.TreeType.ACACIA; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.BIRCH) || feature.is(TreeFeatures.BIRCH_BEES_005)) {
+            treeType = org.bukkit.TreeType.BIRCH; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.SUPER_BIRCH_BEES_0002)) {
+            treeType = org.bukkit.TreeType.TALL_BIRCH; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.SWAMP_OAK)) {
+            treeType = org.bukkit.TreeType.SWAMP; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.FANCY_OAK) || feature.is(TreeFeatures.FANCY_OAK_BEES_005)) {
+            treeType = org.bukkit.TreeType.BIG_TREE; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.JUNGLE_BUSH)) {
+            treeType = org.bukkit.TreeType.JUNGLE_BUSH; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.DARK_OAK)) {
+            treeType = org.bukkit.TreeType.DARK_OAK; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.MEGA_SPRUCE)) {
+            treeType = org.bukkit.TreeType.MEGA_REDWOOD; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.MEGA_PINE)) {
+            treeType = org.bukkit.TreeType.MEGA_PINE; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.MEGA_JUNGLE_TREE)) {
+            treeType = org.bukkit.TreeType.JUNGLE; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.AZALEA_TREE)) {
+            treeType = org.bukkit.TreeType.AZALEA; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.MANGROVE)) {
+            treeType = org.bukkit.TreeType.MANGROVE; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.TALL_MANGROVE)) {
+            treeType = org.bukkit.TreeType.TALL_MANGROVE; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.CHERRY) || feature.is(TreeFeatures.CHERRY_BEES_005)) {
+            treeType = org.bukkit.TreeType.CHERRY; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.PALE_OAK) || feature.is(TreeFeatures.PALE_OAK_BONEMEAL)) {
+            treeType = org.bukkit.TreeType.PALE_OAK; // SparklyPaper - parallel world ticking
+        } else if (feature.is(TreeFeatures.PALE_OAK_CREAKING)) {
+            treeType = org.bukkit.TreeType.PALE_OAK_CREAKING; // SparklyPaper - parallel world ticking
+        } else {
+            throw new IllegalArgumentException("Unknown tree generator " + feature);
+        }
+        net.minecraft.world.level.block.SaplingBlock.treeTypeRT.set(treeType); // SparklyPaper - parallel world ticking
+    }
+    // CraftBukkit end
 }
