--- a/net/minecraft/world/entity/npc/ClientSideMerchant.java
+++ b/net/minecraft/world/entity/npc/ClientSideMerchant.java
@@ -7,6 +_,7 @@
 import net.minecraft.world.item.trading.Merchant;
 import net.minecraft.world.item.trading.MerchantOffer;
 import net.minecraft.world.item.trading.MerchantOffers;
+import org.bukkit.craftbukkit.inventory.CraftMerchant;
 import org.jspecify.annotations.Nullable;
 
 public class ClientSideMerchant implements Merchant {
@@ -54,6 +_,11 @@
     @Override
     public boolean stillValid(Player player) {
         return this.source == player;
+    }
+
+    @Override
+    public CraftMerchant getCraftMerchant() {
+        return null;
     }
 
     @Override
