--- a/net/minecraft/world/entity/npc/InventoryCarrier.java
+++ b/net/minecraft/world/entity/npc/InventoryCarrier.java
@@ -7,6 +_,7 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
+import org.bukkit.event.entity.EntityRemoveEvent;
 
 public interface InventoryCarrier {
     String TAG_INVENTORY = "Inventory";
@@ -22,12 +_,19 @@
                 return;
             }
 
+            // CraftBukkit start
+            ItemStack remaining = new SimpleContainer(inventory).addItem(itemStack);
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(mob, itemEntity, remaining.getCount(), false).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+
             mob.onItemPickup(itemEntity);
             int count = itemStack.getCount();
             ItemStack remainder = inventory.addItem(itemStack);
             mob.take(itemEntity, count - remainder.getCount());
             if (remainder.isEmpty()) {
-                itemEntity.discard();
+                itemEntity.discard(EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
             } else {
                 itemStack.setCount(remainder.getCount());
             }
