--- a/net/minecraft/world/entity/npc/VillagerTrades.java
+++ b/net/minecraft/world/entity/npc/VillagerTrades.java
@@ -1398,7 +_,7 @@
         return PotionContents.createItemStack(Items.POTION, p_316745_);
     }
 
-    public static class DyedArmorForEmeralds implements VillagerTrades.ItemListing {
+    static class DyedArmorForEmeralds implements VillagerTrades.ItemListing {
         private final Item item;
         private final int value;
         private final int maxUses;
@@ -1441,7 +_,7 @@
         }
     }
 
-    public static class EmeraldForItems implements VillagerTrades.ItemListing {
+    static class EmeraldForItems implements VillagerTrades.ItemListing {
         private final ItemCost itemStack;
         private final int maxUses;
         private final int villagerXp;
@@ -1470,13 +_,14 @@
         }
     }
 
-    public static class EmeraldsForVillagerTypeItem implements VillagerTrades.ItemListing {
+    static class EmeraldsForVillagerTypeItem implements VillagerTrades.ItemListing {
         private final Map<VillagerType, Item> trades;
         private final int cost;
         private final int maxUses;
         private final int villagerXp;
 
         public EmeraldsForVillagerTypeItem(int p_35669_, int p_35670_, int p_35671_, Map<VillagerType, Item> p_35672_) {
+            if (false) // Neo: disable this check so that mods can add custom villager types
             BuiltInRegistries.VILLAGER_TYPE.stream().filter(p_35680_ -> !p_35672_.containsKey(p_35680_)).findAny().ifPresent(p_339515_ -> {
                 throw new IllegalStateException("Missing trade for villager type: " + BuiltInRegistries.VILLAGER_TYPE.getKey(p_339515_));
             });
@@ -1490,7 +_,9 @@
         @Override
         public MerchantOffer getOffer(Entity p_219685_, RandomSource p_219686_) {
             if (p_219685_ instanceof VillagerDataHolder villagerdataholder) {
-                ItemCost itemcost = new ItemCost(this.trades.get(villagerdataholder.getVillagerData().getType()), this.cost);
+                Item item = this.trades.get(villagerdataholder.getVillagerData().getType());
+                if (item == null) return null;  // Neo: add a check for unknown villager types
+                ItemCost itemcost = new ItemCost(item, this.cost);
                 return new MerchantOffer(itemcost, new ItemStack(Items.EMERALD), this.maxUses, this.villagerXp, 0.05F);
             } else {
                 return null;
@@ -1498,7 +_,7 @@
         }
     }
 
-    public static class EnchantBookForEmeralds implements VillagerTrades.ItemListing {
+    static class EnchantBookForEmeralds implements VillagerTrades.ItemListing {
         private final int villagerXp;
         private final TagKey<Enchantment> tradeableEnchantments;
         private final int minLevel;
@@ -1547,7 +_,7 @@
         }
     }
 
-    public static class EnchantedItemForEmeralds implements VillagerTrades.ItemListing {
+    static class EnchantedItemForEmeralds implements VillagerTrades.ItemListing {
         private final ItemStack itemStack;
         private final int baseEmeraldCost;
         private final int maxUses;
@@ -1579,8 +_,7 @@
         }
     }
 
-    public static class FailureItemListing implements VillagerTrades.ItemListing {
-        FailureItemListing() {}
+    static class FailureItemListing implements VillagerTrades.ItemListing {
         @Override
         public MerchantOffer getOffer(Entity p_302036_, RandomSource p_301986_) {
             return null;
@@ -1592,7 +_,7 @@
         MerchantOffer getOffer(Entity p_219693_, RandomSource p_219694_);
     }
 
-    public static class ItemsAndEmeraldsToItems implements VillagerTrades.ItemListing {
+    static class ItemsAndEmeraldsToItems implements VillagerTrades.ItemListing {
         private final ItemCost fromItem;
         private final int emeraldCost;
         private final ItemStack toItem;
@@ -1664,7 +_,7 @@
         }
     }
 
-    public static class ItemsForEmeralds implements VillagerTrades.ItemListing {
+    static class ItemsForEmeralds implements VillagerTrades.ItemListing {
         private final ItemStack itemStack;
         private final int emeraldCost;
         private final int maxUses;
@@ -1738,7 +_,7 @@
         }
     }
 
-    public static class SuspiciousStewForEmerald implements VillagerTrades.ItemListing {
+    static class SuspiciousStewForEmerald implements VillagerTrades.ItemListing {
         private final SuspiciousStewEffects effects;
         private final int xp;
         private final float priceMultiplier;
@@ -1762,7 +_,7 @@
         }
     }
 
-    public static class TippedArrowForItemsAndEmeralds implements VillagerTrades.ItemListing {
+    static class TippedArrowForItemsAndEmeralds implements VillagerTrades.ItemListing {
         private final ItemStack toItem;
         private final int toCount;
         private final int emeraldCost;
@@ -1799,7 +_,7 @@
         }
     }
 
-    public static class TreasureMapForEmeralds implements VillagerTrades.ItemListing {
+    static class TreasureMapForEmeralds implements VillagerTrades.ItemListing {
         private final int emeraldCost;
         private final TagKey<Structure> destination;
         private final String displayName;
@@ -1825,7 +_,8 @@
                 return null;
             } else {
                 ServerLevel serverlevel = (ServerLevel)p_219708_.level();
-                BlockPos blockpos = serverlevel.findNearestMapStructure(this.destination, p_219708_.blockPosition(), 100, true);
+                if (!serverlevel.paperConfig().environment.treasureMaps.enabled) return null; // Paper - Configurable cartographer treasure maps
+                BlockPos blockpos = serverlevel.findNearestMapStructure(this.destination, p_219708_.blockPosition(), 100, !serverlevel.paperConfig().environment.treasureMaps.findAlreadyDiscoveredVillager); // Paper - Configurable cartographer treasure maps
                 if (blockpos != null) {
                     ItemStack itemstack = MapItem.create(serverlevel, blockpos.getX(), blockpos.getZ(), (byte)2, true, true);
                     MapItem.renderBiomePreviewMap(serverlevel, itemstack);
@@ -1841,7 +_,7 @@
         }
     }
 
-    public static record TypeSpecificTrade(Map<VillagerType, VillagerTrades.ItemListing> trades) implements VillagerTrades.ItemListing {
+    static record TypeSpecificTrade(Map<VillagerType, VillagerTrades.ItemListing> trades) implements VillagerTrades.ItemListing {
         public static VillagerTrades.TypeSpecificTrade oneTradeInBiomes(VillagerTrades.ItemListing p_302030_, VillagerType... p_301996_) {
             return new VillagerTrades.TypeSpecificTrade(
                 Arrays.stream(p_301996_).collect(Collectors.toMap(p_301920_ -> (VillagerType)p_301920_, p_301922_ -> p_302030_))
