--- a/net/minecraft/world/entity/npc/WanderingTrader.java
+++ b/net/minecraft/world/entity/npc/WanderingTrader.java
@@ -45,6 +_,7 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang3.tuple.Pair;
+import org.bukkit.craftbukkit.inventory.CraftMerchantRecipe;
 
 public class WanderingTrader extends AbstractVillager {
     private static final int NUMBER_OF_TRADE_OFFERS = 5;
@@ -144,7 +_,16 @@
                 VillagerTrades.ItemListing villagertrades$itemlisting = avillagertrades$itemlisting1[i];
                 MerchantOffer merchantoffer = villagertrades$itemlisting.getOffer(this, this.random);
                 if (merchantoffer != null) {
-                    merchantoffers.add(merchantoffer);
+                    // CraftBukkit start
+                    org.bukkit.event.entity.VillagerAcquireTradeEvent event = new org.bukkit.event.entity.VillagerAcquireTradeEvent((org.bukkit.entity.AbstractVillager) getBukkitEntity(), merchantoffer.asBukkit());
+                    // Suppress during worldgen
+                    if (this.valid) {
+                        org.bukkit.Bukkit.getPluginManager().callEvent(event);
+                    }
+                    if (!event.isCancelled()) {
+                        merchantoffers.add(CraftMerchantRecipe.fromBukkit(event.getRecipe()).toMinecraft());
+                    }
+                    // CraftBukkit end
                 }
             }
         }
@@ -240,7 +_,7 @@
 
     private void maybeDespawn() {
         if (this.despawnDelay > 0 && !this.isTrading() && --this.despawnDelay == 0) {
-            this.discard();
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         }
     }
 
