--- a/net/minecraft/world/level/levelgen/PhantomSpawner.java
+++ b/net/minecraft/world/level/levelgen/PhantomSpawner.java
@@ -61,7 +61,7 @@
                                                     if (phantom != null) {
                                                         phantom.snapTo(blockpos1, 0.0F, 0.0F);
                                                         spawngroupdata = phantom.finalizeSpawn(serverlevel, difficultyinstance, EntitySpawnReason.NATURAL, spawngroupdata);
-                                                        serverlevel.addFreshEntityWithPassengers(phantom);
+                                                        serverlevel.addFreshEntityWithPassengers(phantom, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit
                                                     }
                                                 }
                                             }
