--- a/net/minecraft/client/renderer/item/CuboidItemModelWrapper.java
+++ b/net/minecraft/client/renderer/item/CuboidItemModelWrapper.java
@@ -42,7 +_,7 @@
     private final ModelRenderProperties properties;
     private final Matrix4fc transformation;
 
-    public CuboidItemModelWrapper(List<ItemTintSource> tints, QuadCollection quads, ModelRenderProperties properties, Matrix4fc transformation) {
+    private CuboidItemModelWrapper(List<ItemTintSource> tints, QuadCollection quads, ModelRenderProperties properties, Matrix4fc transformation) {
         this.tints = tints;
         this.quads = quads;
         this.properties = properties;
