--- a/net/minecraft/client/gui/render/pip/GuiBookModelRenderer.java
+++ b/net/minecraft/client/gui/render/pip/GuiBookModelRenderer.java
@@ -21,12 +_,12 @@
 
     protected void renderToTexture(GuiBookModelRenderState bookModelState, PoseStack poseStack, SubmitNodeCollector submitNodeCollector) {
         Minecraft.getInstance().gameRenderer.lighting().setupFor(Lighting.Entry.ENTITY_IN_UI);
-        poseStack.rotateDegrees(Axis.YP, 180.0F);
-        poseStack.rotateDegrees(Axis.XP, 25.0F);
+        poseStack.mulPose(Axis.YP.rotationDegrees(180.0F));
+        poseStack.mulPose(Axis.XP.rotationDegrees(25.0F));
         float open = bookModelState.open();
         poseStack.translate((1.0F - open) * 0.2F, (1.0F - open) * 0.1F, (1.0F - open) * 0.25F);
-        poseStack.rotateDegrees(Axis.YP, -(1.0F - open) * 90.0F - 90.0F);
-        poseStack.rotateDegrees(Axis.XP, 180.0F);
+        poseStack.mulPose(Axis.YP.rotationDegrees(-(1.0F - open) * 90.0F - 90.0F));
+        poseStack.mulPose(Axis.XP.rotationDegrees(180.0F));
         float flip = bookModelState.flip();
         float pageFlip1 = Mth.clamp(Mth.frac(flip + 0.25F) * 1.6F - 0.3F, 0.0F, 1.0F);
         float pageFlip2 = Mth.clamp(Mth.frac(flip + 0.75F) * 1.6F - 0.3F, 0.0F, 1.0F);
