--- a/net/minecraft/client/gui/render/pip/GuiEntityRenderer.java
+++ b/net/minecraft/client/gui/render/pip/GuiEntityRenderer.java
@@ -30,7 +_,7 @@
         Minecraft.getInstance().gameRenderer.lighting().setupFor(Lighting.Entry.ENTITY_IN_UI);
         Vector3fc translation = entityState.translation();
         poseStack.translate(translation.x(), translation.y(), translation.z());
-        poseStack.rotate(entityState.rotation());
+        poseStack.mulPose(entityState.rotation());
         Quaternionfc overriddenCameraAngle = entityState.overrideCameraAngle();
         CameraRenderState cameraRenderState = new CameraRenderState();
         if (overriddenCameraAngle != null) {
