--- a/net/minecraft/client/gui/render/pip/GuiSkinRenderer.java
+++ b/net/minecraft/client/gui/render/pip/GuiSkinRenderer.java
@@ -24,7 +_,7 @@
         int guiScale = Minecraft.getInstance().gameRenderer.gameRenderState().windowRenderState.guiScale;
         float scale = skinState.scale() * guiScale;
         RenderSystem.getModelViewStack().rotateAround(Axis.XP.rotationDegrees(skinState.rotationX()), 0.0F, scale * -skinState.pivotY(), 0.0F);
-        modelStack.rotateDegrees(Axis.YP, -skinState.rotationY());
+        modelStack.mulPose(Axis.YP.rotationDegrees(-skinState.rotationY()));
         modelStack.translate(0.0F, -1.6010001F, 0.0F);
         submitNodeCollector.submitModel(skinState.playerModel(), Unit.INSTANCE, modelStack, skinState.texture(), 15728880, OverlayTexture.NO_OVERLAY, 0);
     }
