--- a/net/minecraft/world/item/enchantment/Enchantments.java
+++ b/net/minecraft/world/item/enchantment/Enchantments.java
@@ -82,7 +_,7 @@
 import net.minecraft.world.level.storage.loot.predicates.LootItemRandomChanceCondition;
 import net.minecraft.world.level.storage.loot.predicates.MatchTool;
 import net.minecraft.world.level.storage.loot.predicates.WeatherCheck;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProviders;
+import net.minecraft.world.level.storage.loot.providers.number.EnchantmentLevelProvider;
 import net.minecraft.world.phys.Vec3;
 
 public class Enchantments {
@@ -347,7 +_,7 @@
                         new DamageEntity(LevelBasedValue.constant(1.0F), LevelBasedValue.constant(5.0F), damageTypes.getOrThrow(DamageTypes.THORNS)),
                         new ChangeItemDamage(LevelBasedValue.constant(2.0F))
                     ),
-                    LootItemRandomChanceCondition.randomChance(ContextFloatProviders.forEnchantmentLevel(LevelBasedValue.perLevel(0.15F)))
+                    LootItemRandomChanceCondition.randomChance(EnchantmentLevelProvider.forEnchantmentLevel(LevelBasedValue.perLevel(0.15F)))
                 )
         );
         register(
@@ -413,7 +_,7 @@
                                 BlockPredicate.unobstructed()
                             )
                         ),
-                        BlockStateProvider.holderOf(Blocks.FROSTED_ICE),
+                        BlockStateProvider.simple(Blocks.FROSTED_ICE),
                         Optional.of(GameEvent.BLOCK_PLACE)
                     ),
                     AllOfCondition.allOf(
@@ -529,7 +_,7 @@
                     EnchantmentEffectComponents.LOCATION_CHANGED,
                     new ChangeItemDamage(LevelBasedValue.constant(1.0F)),
                     AllOfCondition.allOf(
-                        LootItemRandomChanceCondition.randomChance(ContextFloatProviders.forEnchantmentLevel(LevelBasedValue.constant(0.04F))),
+                        LootItemRandomChanceCondition.randomChance(EnchantmentLevelProvider.forEnchantmentLevel(LevelBasedValue.constant(0.04F))),
                         LootItemEntityPropertyCondition.hasProperties(
                             LootContext.EntityTarget.THIS,
                             EntityPredicate.Builder.entity()
