--- a/net/minecraft/world/item/enchantment/ItemEnchantments.java
+++ b/net/minecraft/world/item/enchantment/ItemEnchantments.java
@@ -49,7 +_,7 @@
         ItemEnchantments::new
     );
     final Object2IntOpenHashMap<Holder<Enchantment>> enchantments;
-    public final boolean showInTooltip;
+    final boolean showInTooltip;
 
     ItemEnchantments(Object2IntOpenHashMap<Holder<Enchantment>> p_341287_, boolean p_330219_) {
         this.enchantments = p_341287_;
@@ -144,7 +_,7 @@
 
     public static class Mutable {
         private final Object2IntOpenHashMap<Holder<Enchantment>> enchantments = new Object2IntOpenHashMap<>();
-        public boolean showInTooltip;
+        private final boolean showInTooltip;
 
         public Mutable(ItemEnchantments p_330722_) {
             this.enchantments.putAll(p_330722_.enchantments);
