--- a/net/minecraft/world/item/enchantment/LevelBasedValue.java
+++ b/net/minecraft/world/item/enchantment/LevelBasedValue.java
@@ -47,7 +_,7 @@
 
     MapCodec<? extends LevelBasedValue> codec();
 
-    record Clamped(LevelBasedValue value, float min, float max) implements LevelBasedValue {
+    public record Clamped(LevelBasedValue value, float min, float max) implements LevelBasedValue {
         public static final MapCodec<LevelBasedValue.Clamped> CODEC = RecordCodecBuilder.<LevelBasedValue.Clamped>mapCodec(
                 i -> i.group(
                         LevelBasedValue.CODEC.fieldOf("value").forGetter(LevelBasedValue.Clamped::value),
@@ -69,7 +_,7 @@
         }
     }
 
-    record Constant(float value) implements LevelBasedValue {
+    public record Constant(float value) implements LevelBasedValue {
         public static final Codec<LevelBasedValue.Constant> CODEC = Codec.FLOAT.xmap(LevelBasedValue.Constant::new, LevelBasedValue.Constant::value);
         public static final MapCodec<LevelBasedValue.Constant> TYPED_CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(Codec.FLOAT.fieldOf("value").forGetter(LevelBasedValue.Constant::value)).apply(i, LevelBasedValue.Constant::new)
@@ -86,7 +_,7 @@
         }
     }
 
-    record Exponent(LevelBasedValue base, LevelBasedValue power) implements LevelBasedValue {
+    public record Exponent(LevelBasedValue base, LevelBasedValue power) implements LevelBasedValue {
         public static final MapCodec<LevelBasedValue.Exponent> CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     LevelBasedValue.CODEC.fieldOf("base").forGetter(LevelBasedValue.Exponent::base),
@@ -106,7 +_,7 @@
         }
     }
 
-    record Fraction(LevelBasedValue numerator, LevelBasedValue denominator) implements LevelBasedValue {
+    public record Fraction(LevelBasedValue numerator, LevelBasedValue denominator) implements LevelBasedValue {
         public static final MapCodec<LevelBasedValue.Fraction> CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     LevelBasedValue.CODEC.fieldOf("numerator").forGetter(LevelBasedValue.Fraction::numerator),
@@ -127,7 +_,7 @@
         }
     }
 
-    record LevelsSquared(float added) implements LevelBasedValue {
+    public record LevelsSquared(float added) implements LevelBasedValue {
         public static final MapCodec<LevelBasedValue.LevelsSquared> CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(Codec.FLOAT.fieldOf("added").forGetter(LevelBasedValue.LevelsSquared::added)).apply(i, LevelBasedValue.LevelsSquared::new)
         );
@@ -143,7 +_,7 @@
         }
     }
 
-    record Linear(float base, float perLevelAboveFirst) implements LevelBasedValue {
+    public record Linear(float base, float perLevelAboveFirst) implements LevelBasedValue {
         public static final MapCodec<LevelBasedValue.Linear> CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     Codec.FLOAT.fieldOf("base").forGetter(LevelBasedValue.Linear::base),
@@ -163,7 +_,7 @@
         }
     }
 
-    record Lookup(List<Float> values, LevelBasedValue fallback) implements LevelBasedValue {
+    public record Lookup(List<Float> values, LevelBasedValue fallback) implements LevelBasedValue {
         public static final MapCodec<LevelBasedValue.Lookup> CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     Codec.FLOAT.listOf().fieldOf("values").forGetter(LevelBasedValue.Lookup::values),
