--- a/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
+++ b/net/minecraft/network/protocol/common/custom/DiscardedPayload.java
@@ -4,13 +_,31 @@
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.resources.Identifier;
 
-public record DiscardedPayload(Identifier id) implements CustomPacketPayload {
+public class DiscardedPayload implements CustomPacketPayload { // Paper - store data
+    private final Identifier id;
+    private byte[] data;
+
+    public DiscardedPayload(Identifier id) {
+        this.id = id;
+    }
+
+    public DiscardedPayload(Identifier id, byte[] data) {
+        this.id = id;
+        this.data = data;
+    }
+
     public static <T extends FriendlyByteBuf> StreamCodec<T, DiscardedPayload> codec(Identifier id, int maxPayloadSize) {
-        return CustomPacketPayload.codec((payload, buf) -> {}, buf -> {
+        return CustomPacketPayload.codec((payload, buf) -> {
+            // Paper start
+            // Always write data
+            buf.writeBytes(payload.data);
+        }, buf -> {
             int length = buf.readableBytes();
             if (length >= 0 && length <= maxPayloadSize) {
-                buf.skipBytes(length);
-                return new DiscardedPayload(id);
+                final byte[] data = new byte[length];
+                buf.readBytes(data);
+                return new DiscardedPayload(id, data);
+                // Paper end
             } else {
                 throw new IllegalArgumentException("Payload may not be larger than " + maxPayloadSize + " bytes");
             }
@@ -20,5 +_,13 @@
     @Override
     public CustomPacketPayload.Type<DiscardedPayload> type() {
         return new CustomPacketPayload.Type<>(this.id);
+    }
+
+    public Identifier id() {
+        return id;
+    }
+
+    public byte[] data() {
+        return data;
     }
 }
