--- a/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
@@ -1,5 +_,7 @@
 package net.minecraft.world.entity.vehicle;
 
+import com.destroystokyo.paper.loottable.PaperLootableInventoryData;
+import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.core.NonNullList;
 import net.minecraft.nbt.CompoundTag;
@@ -17,6 +_,10 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.storage.loot.LootTable;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
 
 public abstract class AbstractMinecartContainer extends AbstractMinecart implements ContainerEntity {
     private NonNullList<ItemStack> itemStacks = NonNullList.withSize(36, ItemStack.EMPTY);
@@ -24,6 +_,56 @@
     public ResourceKey<LootTable> lootTable;
     public long lootTableSeed;
 
+    // Paper start - LootTable API
+    final PaperLootableInventoryData lootableData = new PaperLootableInventoryData();
+
+    @Override
+    public PaperLootableInventoryData lootableData() {
+        return this.lootableData;
+    }
+    // Paper end - LootTable API
+
+    // CraftBukkit start
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+
+    public List<ItemStack> getContents() {
+        return this.itemStacks;
+    }
+
+    public void onOpen(CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    public InventoryHolder getOwner() {
+        org.bukkit.entity.Entity cart = getBukkitEntity();
+        if(cart instanceof InventoryHolder) return (InventoryHolder) cart;
+        return null;
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return maxStack;
+    }
+
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+
+    @Override
+    public Location getLocation() {
+        return getBukkitEntity().getLocation();
+    }
+    // CraftBukkit end
+
     protected AbstractMinecartContainer(EntityType<?> p_38213_, Level p_38214_) {
         super(p_38213_, p_38214_);
     }
@@ -74,11 +_,18 @@
 
     @Override
     public void remove(Entity.RemovalReason p_150255_) {
+        // CraftBukkit start - add Bukkit remove cause
+        this.removeCB(p_150255_, null);
+    }
+
+    @Override
+    public void removeCB(Entity.RemovalReason p_150255_, org.bukkit.event.entity.EntityRemoveEvent.Cause cause) {
+        // CraftBukkit end
         if (!this.level().isClientSide && p_150255_.shouldDestroy()) {
             Containers.dropContents(this.level(), this, this);
         }
 
-        super.remove(p_150255_);
+        super.removeCB(p_150255_, cause); // CraftBukkit - add Bukkit remove cause
     }
 
     @Override
@@ -95,6 +_,8 @@
 
     @Override
     public InteractionResult interact(Player p_38232_, InteractionHand p_38233_) {
+        InteractionResult ret = super.interact(p_38232_, p_38233_);
+        if (ret.consumesAction()) return ret;
         return this.interactWithContainerVehicle(p_38232_);
     }
 
