--- a/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
@@ -21,8 +_,8 @@
 public abstract class AbstractMinecartContainer extends AbstractMinecart implements ContainerEntity {
     private NonNullList<ItemStack> itemStacks = NonNullList.withSize(36, ItemStack.EMPTY);
     @Nullable
-    public ResourceKey<LootTable> lootTable;
-    public long lootTableSeed;
+    private ResourceKey<LootTable> lootTable;
+    private long lootTableSeed;
 
     protected AbstractMinecartContainer(EntityType<?> p_38213_, Level p_38214_) {
         super(p_38213_, p_38214_);
@@ -95,6 +_,8 @@
 
     @Override
     public InteractionResult interact(Player p_38232_, InteractionHand p_38233_) {
+        InteractionResult ret = super.interact(p_38232_, p_38233_);
+        if (ret.consumesAction()) return ret;
         return this.interactWithContainerVehicle(p_38232_);
     }
 
