--- a/net/minecraft/world/entity/vehicle/ChestBoat.java
+++ b/net/minecraft/world/entity/vehicle/ChestBoat.java
@@ -1,5 +_,7 @@
 package net.minecraft.world.entity.vehicle;
 
+import com.destroystokyo.paper.loottable.PaperLootableInventoryData;
+import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.core.NonNullList;
 import net.minecraft.nbt.CompoundTag;
@@ -23,6 +_,10 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.storage.loot.LootTable;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
 
 public class ChestBoat extends Boat implements HasCustomInventoryScreen, ContainerEntity {
     private static final int CONTAINER_SIZE = 27;
@@ -73,11 +_,18 @@
 
     @Override
     public void remove(Entity.RemovalReason p_219894_) {
+        // CraftBukkit start - add Bukkit remove cause
+        this.removeCB(p_219894_, null);
+    }
+
+    @Override
+    public void removeCB(Entity.RemovalReason p_219894_, org.bukkit.event.entity.EntityRemoveEvent.Cause cause) {
+        // CraftBukkit end
         if (!this.level().isClientSide && p_219894_.shouldDestroy()) {
             Containers.dropContents(this.level(), this, this);
         }
 
-        super.remove(p_219894_);
+        super.removeCB(p_219894_, cause); // CraftBukkit - add Bukkit remove cause
     }
 
     @Override
@@ -122,7 +_,8 @@
             case DARK_OAK -> Items.DARK_OAK_CHEST_BOAT;
             case MANGROVE -> Items.MANGROVE_CHEST_BOAT;
             case BAMBOO -> Items.BAMBOO_CHEST_RAFT;
-            default -> Items.OAK_CHEST_BOAT;
+            case OAK -> Items.OAK_CHEST_BOAT;
+            default -> this.getVariant().chestBoatItem.get();
         };
     }
 
@@ -220,4 +_,59 @@
     public void stopOpen(Player p_270286_) {
         this.level().gameEvent(GameEvent.CONTAINER_CLOSE, this.position(), GameEvent.Context.of(p_270286_));
     }
+
+    // Paper start - LootTable API
+    final PaperLootableInventoryData lootableData = new PaperLootableInventoryData();
+
+    @Override
+    public PaperLootableInventoryData lootableData() {
+        return this.lootableData;
+    }
+    // Paper end - LootTable API
+    // CraftBukkit start
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+
+    @Override
+    public List<ItemStack> getContents() {
+        return this.itemStacks;
+    }
+
+    @Override
+    public void onOpen(CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    @Override
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    @Override
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    @Override
+    public InventoryHolder getOwner() {
+        org.bukkit.entity.Entity entity = getBukkitEntity();
+        if (entity instanceof InventoryHolder) return (InventoryHolder) entity;
+        return null;
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return maxStack;
+    }
+
+    @Override
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+
+    @Override
+    public Location getLocation() {
+        return getBukkitEntity().getLocation();
+    }
+    // CraftBukkit end
 }
