--- a/net/minecraft/world/entity/vehicle/ContainerEntity.java
+++ b/net/minecraft/world/entity/vehicle/ContainerEntity.java
@@ -100,6 +_,9 @@
 
             this.setContainerLootTable(null);
             LootParams.Builder builder = new LootParams.Builder((ServerLevel)this.level()).withParameter(LootContextParams.ORIGIN, this.position());
+            // Neo: set the chest to attacking_entity for loot context.
+            if (this instanceof net.minecraft.world.entity.vehicle.minecart.AbstractMinecartContainer entityContainer)
+                builder.withParameter(LootContextParams.ATTACKING_ENTITY, entityContainer);
             if (player != null) {
                 builder.withLuck(player.getLuck()).withParameter(LootContextParams.THIS_ENTITY, player);
             }
@@ -168,4 +_,14 @@
     default boolean isChestVehicleStillValid(Player player) {
         return !this.isRemoved() && player.isWithinEntityInteractionRange(this.getBoundingBox(), 4.0);
     }
+
+    // Paper start - LootTable API
+    default com.destroystokyo.paper.loottable.PaperLootableInventoryData lootableData() {
+        throw new UnsupportedOperationException("Implement this method");
+    }
+
+    default com.destroystokyo.paper.loottable.PaperLootableInventory getLootableInventory() {
+        return ((com.destroystokyo.paper.loottable.PaperLootableInventory) ((Entity) this).getBukkitEntity());
+    }
+    // Paper end - LootTable API
 }
