--- a/net/minecraft/world/entity/vehicle/MinecartTNT.java
+++ b/net/minecraft/world/entity/vehicle/MinecartTNT.java
@@ -22,7 +_,7 @@
 
 public class MinecartTNT extends AbstractMinecart {
     private static final byte EVENT_PRIME = 10;
-    public int fuse = -1;
+    private int fuse = -1;
 
     public MinecartTNT(EntityType<? extends MinecartTNT> p_38649_, Level p_38650_) {
         super(p_38649_, p_38650_);
@@ -88,11 +_,11 @@
         return Items.TNT_MINECART;
     }
 
-    public void explode(double p_38689_) {
+    protected void explode(double p_38689_) {
         this.explode(null, p_38689_);
     }
 
-    public void explode(@Nullable DamageSource p_259539_, double p_260287_) {
+    protected void explode(@Nullable DamageSource p_259539_, double p_260287_) {
         if (!this.level().isClientSide) {
             double d0 = Math.sqrt(p_260287_);
             if (d0 > 5.0) {
