--- a/net/minecraft/world/entity/vehicle/MinecartTNT.java
+++ b/net/minecraft/world/entity/vehicle/MinecartTNT.java
@@ -19,6 +_,7 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
+import org.bukkit.event.entity.EntityRemoveEvent;
 
 public class MinecartTNT extends AbstractMinecart {
     private static final byte EVENT_PRIME = 10;
@@ -46,6 +_,12 @@
     public void tick() {
         super.tick();
         if (this.fuse > 0) {
+            // Paper start - Configurable TNT height nerf
+            if (this.level().paperConfig().fixes.tntEntityHeightNerf.test(v -> this.getY() > v)) {
+                this.discard(EntityRemoveEvent.Cause.OUT_OF_WORLD);
+                return;
+            }
+            // Paper end - Configurable TNT height nerf
             this.fuse--;
             this.level().addParticle(ParticleTypes.SMOKE, this.getX(), this.getY() + 0.5, this.getZ(), 0.0, 0.0, 0.0);
         } else if (this.fuse == 0) {
@@ -92,26 +_,23 @@
         this.explode(null, p_38689_);
     }
 
-    public void explode(@Nullable DamageSource p_259539_, double p_260287_) {
+    protected void explode(@Nullable DamageSource p_259539_, double p_260287_) {
         if (!this.level().isClientSide) {
             double d0 = Math.sqrt(p_260287_);
             if (d0 > 5.0) {
                 d0 = 5.0;
             }
 
-            this.level()
-                .explode(
-                    this,
-                    p_259539_,
-                    null,
-                    this.getX(),
-                    this.getY(),
-                    this.getZ(),
-                    (float)(4.0 + this.random.nextDouble() * 1.5 * d0),
-                    false,
-                    Level.ExplosionInteraction.TNT
-                );
-            this.discard();
+            // CraftBukkit start
+            org.bukkit.event.entity.ExplosionPrimeEvent event = new org.bukkit.event.entity.ExplosionPrimeEvent(this.getBukkitEntity(), (float) (4.0D + this.random.nextDouble() * 1.5D * d0), false);
+            this.level().getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                fuse = -1;
+                return;
+            }
+            this.level().explode(this, p_259539_, (net.minecraft.world.level.ExplosionDamageCalculator) null, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.TNT);
+            // CraftBukkit end
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit - add Bukkit remove cause
         }
     }
 
