--- a/net/minecraft/world/entity/vehicle/VehicleEntity.java
+++ b/net/minecraft/world/entity/vehicle/VehicleEntity.java
@@ -16,6 +_,10 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.gamerules.GameRules;
+import org.bukkit.entity.Vehicle;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.vehicle.VehicleDamageEvent;
+import org.bukkit.event.vehicle.VehicleDestroyEvent;
 
 public abstract class VehicleEntity extends Entity {
     protected static final EntityDataAccessor<Integer> DATA_ID_HURT = SynchedEntityData.defineId(VehicleEntity.class, EntityDataSerializers.INT);
@@ -35,27 +_,55 @@
     public boolean hurtServer(ServerLevel level, DamageSource source, float damage) {
         if (this.isRemoved()) {
             return true;
-        }
-
-        if (this.isInvulnerableToBase(source)) {
+        } else if (this.isInvulnerableToBase(source)) {
             return false;
-        }
-
-        this.setHurtDir(-this.getHurtDir());
-        this.setHurtTime(10);
-        this.markHurt();
-        this.setDamage(this.getDamage() + damage * 10.0F);
-        this.gameEvent(GameEvent.ENTITY_DAMAGE, source.getEntity());
-        boolean creativePlayer = source.getEntity() instanceof Player player && player.getAbilities().instabuild;
-        if ((creativePlayer || !(this.getDamage() > 40.0F)) && !this.shouldSourceDestroy(source)) {
-            if (creativePlayer) {
-                this.discard();
-            }
         } else {
-            this.destroy(level, source);
+            this.setHurtDir(-this.getHurtDir());
+            this.setHurtTime(10);
+            this.markHurt();
+            this.setDamage(this.getDamage() + damage * 10.0F);
+            this.gameEvent(GameEvent.ENTITY_DAMAGE, source.getEntity());
+            // CraftBukkit start
+            Vehicle vehicle = (Vehicle) this.getBukkitEntity();
+            org.bukkit.entity.Entity attacker = (source.getEntity() == null) ? null : source.getEntity().getBukkitEntity();
+
+            VehicleDamageEvent event = new VehicleDamageEvent(vehicle, attacker, (double) damage);
+            this.level().getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return false;
+            }
+            damage = (float) event.getDamage();
+            // CraftBukkit end
+            boolean creativePlayer = source.getEntity() instanceof Player player && player.getAbilities().instabuild;
+            if ((creativePlayer || !(this.getDamage() > 40.0F)) && !this.shouldSourceDestroy(source)) {
+                if (creativePlayer) {
+                    // CraftBukkit start
+                    VehicleDestroyEvent destroyEvent = new VehicleDestroyEvent(vehicle, attacker);
+                    this.level().getCraftServer().getPluginManager().callEvent(destroyEvent);
+
+                    if (destroyEvent.isCancelled()) {
+                        this.setDamage(40.0F);
+                        return true;
+                    }
+                    // CraftBukkit end
+                    this.discard(EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause
+                }
+            } else {
+                // CraftBukkit start
+                VehicleDestroyEvent destroyEvent = new VehicleDestroyEvent(vehicle, attacker);
+                this.level().getCraftServer().getPluginManager().callEvent(destroyEvent);
+
+                if (destroyEvent.isCancelled()) {
+                    this.setDamage(40.0F);
+                    return true;
+                }
+                // CraftBukkit end
+                this.destroy(level, source);
+            }
+
+            return true;
         }
-
-        return true;
     }
 
     protected boolean shouldSourceDestroy(DamageSource source) {
