--- a/net/minecraft/core/cauldron/CauldronInteraction.java
+++ b/net/minecraft/core/cauldron/CauldronInteraction.java
@@ -36,14 +36,20 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.material.FluidState;
 
+// CraftBukkit start
+import org.bukkit.event.block.CauldronLevelChangeEvent;
+// CraftBukkit end
+
 public interface CauldronInteraction {
 
     Map<String, CauldronInteraction.InteractionMap> INTERACTIONS = new Object2ObjectArrayMap();
-    Codec<CauldronInteraction.InteractionMap> CODEC;
-    CauldronInteraction.InteractionMap EMPTY;
-    CauldronInteraction.InteractionMap WATER;
-    CauldronInteraction.InteractionMap LAVA;
-    CauldronInteraction.InteractionMap POWDER_SNOW;
+    // CraftBukkit start - decompile errors
+    Codec<CauldronInteraction.InteractionMap> CODEC = Codec.stringResolver(CauldronInteraction.InteractionMap::name, CauldronInteraction.INTERACTIONS::get);
+    CauldronInteraction.InteractionMap EMPTY = newInteractionMap("empty");
+    CauldronInteraction.InteractionMap WATER = newInteractionMap("water");
+    CauldronInteraction.InteractionMap LAVA = newInteractionMap("lava");
+    CauldronInteraction.InteractionMap POWDER_SNOW = newInteractionMap("powder_snow");
+    // CraftBukkit end
 
     static CauldronInteraction.InteractionMap newInteractionMap(String s) {
         Object2ObjectOpenHashMap<Item, CauldronInteraction> object2objectopenhashmap = new Object2ObjectOpenHashMap();
@@ -68,12 +74,17 @@
 
             if (potioncontents != null && potioncontents.is(Potions.WATER)) {
                 if (!level.isClientSide()) {
+                    // CraftBukkit start
+                    if (!LayeredCauldronBlock.changeLevel(blockstate, level, blockpos, Blocks.WATER_CAULDRON.defaultBlockState(), player, CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                        return InteractionResult.SUCCESS;
+                    }
+                    // CraftBukkit end
                     Item item = itemstack.getItem();
 
                     player.setItemInHand(interactionhand, ItemUtils.createFilledResult(itemstack, player, new ItemStack(Items.GLASS_BOTTLE)));
                     player.awardStat(Stats.USE_CAULDRON);
                     player.awardStat(Stats.ITEM_USED.get(item));
-                    level.setBlockAndUpdate(blockpos, Blocks.WATER_CAULDRON.defaultBlockState());
+                    // world.setBlockAndUpdate(blockposition, Blocks.WATER_CAULDRON.defaultBlockState()); // CraftBukkit
                     level.playSound((Entity) null, blockpos, SoundEvents.BOTTLE_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);
                     level.gameEvent((Entity) null, (Holder) GameEvent.FLUID_PLACE, blockpos);
                 }
@@ -93,12 +104,17 @@
         });
         map1.put(Items.GLASS_BOTTLE, (CauldronInteraction) (blockstate, level, blockpos, player, interactionhand, itemstack) -> {
             if (!level.isClientSide()) {
+                // CraftBukkit start
+                if (!LayeredCauldronBlock.lowerFillLevel(blockstate, level, blockpos, player, CauldronLevelChangeEvent.ChangeReason.BOTTLE_FILL)) {
+                    return InteractionResult.SUCCESS;
+                }
+                // CraftBukkit end
                 Item item = itemstack.getItem();
 
                 player.setItemInHand(interactionhand, ItemUtils.createFilledResult(itemstack, player, PotionContents.createItemStack(Items.POTION, Potions.WATER)));
                 player.awardStat(Stats.USE_CAULDRON);
                 player.awardStat(Stats.ITEM_USED.get(item));
-                LayeredCauldronBlock.lowerFillLevel(blockstate, level, blockpos);
+                // LayeredCauldronBlock.lowerFillLevel(iblockdata, world, blockposition); // CraftBukkit
                 level.playSound((Entity) null, blockpos, SoundEvents.BOTTLE_FILL, SoundSource.BLOCKS, 1.0F, 1.0F);
                 level.gameEvent((Entity) null, (Holder) GameEvent.FLUID_PICKUP, blockpos);
             }
@@ -113,10 +129,15 @@
 
                 if (potioncontents != null && potioncontents.is(Potions.WATER)) {
                     if (!level.isClientSide()) {
+                        // CraftBukkit start
+                        if (!LayeredCauldronBlock.changeLevel(blockstate, level, blockpos, blockstate.cycle(LayeredCauldronBlock.LEVEL), player, CauldronLevelChangeEvent.ChangeReason.BOTTLE_EMPTY)) {
+                            return InteractionResult.SUCCESS;
+                        }
+                        // CraftBukkit end
                         player.setItemInHand(interactionhand, ItemUtils.createFilledResult(itemstack, player, new ItemStack(Items.GLASS_BOTTLE)));
                         player.awardStat(Stats.USE_CAULDRON);
                         player.awardStat(Stats.ITEM_USED.get(itemstack.getItem()));
-                        level.setBlockAndUpdate(blockpos, (BlockState) blockstate.cycle(LayeredCauldronBlock.LEVEL));
+                        // world.setBlockAndUpdate(blockposition, (IBlockData) iblockdata.cycle(LayeredCauldronBlock.LEVEL)); // CraftBukkit
                         level.playSound((Entity) null, blockpos, SoundEvents.BOTTLE_EMPTY, SoundSource.BLOCKS, 1.0F, 1.0F);
                         level.gameEvent((Entity) null, (Holder) GameEvent.FLUID_PLACE, blockpos);
                     }
@@ -194,12 +215,17 @@
             return InteractionResult.TRY_WITH_EMPTY_HAND;
         } else {
             if (!level.isClientSide()) {
+                // CraftBukkit start
+                if (!LayeredCauldronBlock.changeLevel(blockstate, level, blockpos, Blocks.CAULDRON.defaultBlockState(), player, CauldronLevelChangeEvent.ChangeReason.BUCKET_FILL)) {
+                    return InteractionResult.SUCCESS;
+                }
+                // CraftBukkit end
                 Item item = itemstack.getItem();
 
                 player.setItemInHand(interactionhand, ItemUtils.createFilledResult(itemstack, player, itemstack1));
                 player.awardStat(Stats.USE_CAULDRON);
                 player.awardStat(Stats.ITEM_USED.get(item));
-                level.setBlockAndUpdate(blockpos, Blocks.CAULDRON.defaultBlockState());
+                // world.setBlockAndUpdate(blockposition, Blocks.CAULDRON.defaultBlockState()); // CraftBukkit
                 level.playSound((Entity) null, blockpos, soundevent, SoundSource.BLOCKS, 1.0F, 1.0F);
                 level.gameEvent((Entity) null, (Holder) GameEvent.FLUID_PICKUP, blockpos);
             }
@@ -210,12 +236,17 @@
 
     static InteractionResult emptyBucket(Level level, BlockPos blockpos, Player player, InteractionHand interactionhand, ItemStack itemstack, BlockState blockstate, SoundEvent soundevent) {
         if (!level.isClientSide()) {
+            // CraftBukkit start
+            if (!LayeredCauldronBlock.changeLevel(blockstate, level, blockpos, blockstate, player, CauldronLevelChangeEvent.ChangeReason.BUCKET_EMPTY)) {
+                return InteractionResult.SUCCESS;
+            }
+            // CraftBukkit end
             Item item = itemstack.getItem();
 
             player.setItemInHand(interactionhand, ItemUtils.createFilledResult(itemstack, player, new ItemStack(Items.BUCKET)));
             player.awardStat(Stats.FILL_CAULDRON);
             player.awardStat(Stats.ITEM_USED.get(item));
-            level.setBlockAndUpdate(blockpos, blockstate);
+            // world.setBlockAndUpdate(blockposition, iblockdata); // CraftBukkit
             level.playSound((Entity) null, blockpos, soundevent, SoundSource.BLOCKS, 1.0F, 1.0F);
             level.gameEvent((Entity) null, (Holder) GameEvent.FLUID_PLACE, blockpos);
         }
@@ -242,11 +273,16 @@
             return InteractionResult.TRY_WITH_EMPTY_HAND;
         } else {
             if (!level.isClientSide()) {
+                // CraftBukkit start
+                if (!LayeredCauldronBlock.lowerFillLevel(blockstate, level, blockpos, player, CauldronLevelChangeEvent.ChangeReason.SHULKER_WASH)) {
+                    return InteractionResult.SUCCESS;
+                }
+                // CraftBukkit end
                 ItemStack itemstack1 = itemstack.transmuteCopy(Blocks.SHULKER_BOX, 1);
 
                 player.setItemInHand(interactionhand, ItemUtils.createFilledResult(itemstack, player, itemstack1, false));
                 player.awardStat(Stats.CLEAN_SHULKER_BOX);
-                LayeredCauldronBlock.lowerFillLevel(blockstate, level, blockpos);
+                // LayeredCauldronBlock.lowerFillLevel(iblockdata, world, blockposition); // CraftBukkit
             }
 
             return InteractionResult.SUCCESS;
@@ -260,12 +296,17 @@
             return InteractionResult.TRY_WITH_EMPTY_HAND;
         } else {
             if (!level.isClientSide()) {
+                // CraftBukkit start
+                if (!LayeredCauldronBlock.lowerFillLevel(blockstate, level, blockpos, player, CauldronLevelChangeEvent.ChangeReason.BANNER_WASH)) {
+                    return InteractionResult.SUCCESS;
+                }
+                // CraftBukkit end
                 ItemStack itemstack1 = itemstack.copyWithCount(1);
 
                 itemstack1.set(DataComponents.BANNER_PATTERNS, bannerpatternlayers.removeLast());
                 player.setItemInHand(interactionhand, ItemUtils.createFilledResult(itemstack, player, itemstack1, false));
                 player.awardStat(Stats.CLEAN_BANNER);
-                LayeredCauldronBlock.lowerFillLevel(blockstate, level, blockpos);
+                // LayeredCauldronBlock.lowerFillLevel(iblockdata, world, blockposition); // CraftBukkit
             }
 
             return InteractionResult.SUCCESS;
@@ -279,9 +320,14 @@
             return InteractionResult.TRY_WITH_EMPTY_HAND;
         } else {
             if (!level.isClientSide()) {
+                // CraftBukkit start
+                if (!LayeredCauldronBlock.lowerFillLevel(blockstate, level, blockpos, player, CauldronLevelChangeEvent.ChangeReason.ARMOR_WASH)) {
+                    return InteractionResult.SUCCESS;
+                }
+                // CraftBukkit end
                 itemstack.remove(DataComponents.DYED_COLOR);
                 player.awardStat(Stats.CLEAN_ARMOR);
-                LayeredCauldronBlock.lowerFillLevel(blockstate, level, blockpos);
+                // LayeredCauldronBlock.lowerFillLevel(iblockdata, world, blockposition); // CraftBukkit
             }
 
             return InteractionResult.SUCCESS;
@@ -294,6 +340,8 @@
         return fluidstate.is(FluidTags.WATER);
     }
 
+    // CraftBukkit start - decompile errors
+    /*
     static {
         Function function = CauldronInteraction.InteractionMap::name;
         Map map = CauldronInteraction.INTERACTIONS;
@@ -305,6 +353,8 @@
         LAVA = newInteractionMap("lava");
         POWDER_SNOW = newInteractionMap("powder_snow");
     }
+     */
+    // CraftBukkit end
 
     public static record InteractionMap(String name, Map<Item, CauldronInteraction> map) {
 
