--- a/net/minecraft/world/item/consume_effects/ClearAllStatusEffectsConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/ClearAllStatusEffectsConsumeEffect.java
@@ -6,6 +_,7 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
 
 public record ClearAllStatusEffectsConsumeEffect() implements ConsumeEffect {
     public static final ClearAllStatusEffectsConsumeEffect INSTANCE = new ClearAllStatusEffectsConsumeEffect();
@@ -18,7 +_,14 @@
     }
 
     @Override
-    public boolean apply(Level level, ItemStack stack, LivingEntity user) {
+    public boolean apply(Level level, ItemStack stack, LivingEntity user, EntityPotionEffectEvent.Cause cause) {
+        // Purpur start - Option to toggle milk curing bad omen
+        net.minecraft.world.effect.MobEffectInstance badOmen = user.getEffect(net.minecraft.world.effect.MobEffects.BAD_OMEN);
+        if (!level.purpurConfig.milkCuresBadOmen && stack.is(net.minecraft.world.item.Items.MILK_BUCKET) && badOmen != null) {
+            return user.removeAllEffects(cause) && user.addEffect(badOmen);
+        }
+        // Purpur end - Option to toggle milk curing bad omen
+        user.removeAllEffects$cause(cause);
         return user.removeAllEffects();
     }
 }
