--- a/net/minecraft/world/item/consume_effects/RemoveStatusEffectsConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/RemoveStatusEffectsConsumeEffect.java
@@ -33,10 +_,11 @@
     }
 
     @Override
-    public boolean apply(Level level, ItemStack stack, LivingEntity user) {
+    public boolean apply(Level level, ItemStack stack, LivingEntity user, EntityPotionEffectEvent.Cause cause) { // CraftBukkit
         boolean hasRemovedAny = false;
 
         for (Holder<MobEffect> effect : this.effects) {
+            user.removeAllEffects$cause(cause);
             if (user.removeEffect(effect)) {
                 hasRemovedAny = true;
             }
