--- a/net/minecraft/world/item/consume_effects/RemoveStatusEffectsConsumeEffect.java
+++ b/net/minecraft/world/item/consume_effects/RemoveStatusEffectsConsumeEffect.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
 
 public record RemoveStatusEffectsConsumeEffect(HolderSet<MobEffect> effects) implements ConsumeEffect {
     public static final MapCodec<RemoveStatusEffectsConsumeEffect> CODEC = RecordCodecBuilder.mapCodec(
@@ -33,10 +_,11 @@
     }
 
     @Override
-    public boolean apply(Level level, ItemStack stack, LivingEntity user) {
+    public boolean apply(Level level, ItemStack stack, LivingEntity user, EntityPotionEffectEvent.Cause cause) { // CraftBukkit
         boolean hasRemovedAny = false;
 
         for (Holder<MobEffect> effect : this.effects) {
+            user.removeAllEffects$cause(cause);
             if (user.removeEffect(effect)) {
                 hasRemovedAny = true;
             }
