--- a/net/minecraft/client/resources/ClientPackSource.java
+++ b/net/minecraft/client/resources/ClientPackSource.java
@@ -67,7 +_,7 @@
         return null;
     }
 
-    public static VanillaPackResources createVanillaPackSource(Path externalAssetRoot) {
+    private static VanillaPackResources createVanillaPackSource(Path externalAssetRoot) {
         return new VanillaPackResourcesBuilder()
             .setMetadata(BUILT_IN_METADATA)
             .exposeNamespace("minecraft", "realms")
