--- a/net/minecraft/client/resources/SplashManager.java
+++ b/net/minecraft/client/resources/SplashManager.java
@@ -26,7 +_,7 @@
     public static final Component CHRISTMAS = literalSplash("Merry X-mas!");
     public static final Component NEW_YEAR = literalSplash("Happy new year!");
     public static final Component HALLOWEEN = literalSplash("OOoooOOOoooo! Spooky!");
-    public static final Identifier SPLASHES_LOCATION = Identifier.withDefaultNamespace("texts/splashes.txt");
+    private static final Identifier SPLASHES_LOCATION = Identifier.withDefaultNamespace("texts/splashes.txt");
     private static final RandomSource RANDOM = RandomSource.create();
     private List<Component> splashes = List.of();
     private final User user;
@@ -35,7 +_,7 @@
         this.user = user;
     }
 
-    public static Component literalSplash(String text) {
+    private static Component literalSplash(String text) {
         return Component.literal(text).setStyle(DEFAULT_STYLE);
     }
 
