--- a/net/minecraft/core/BlockPos.java
+++ b/net/minecraft/core/BlockPos.java
@@ -27,6 +_,7 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang3.Validate;
+import org.apache.commons.lang3.tuple.Pair;
 
 @Immutable
 public class BlockPos extends Vec3i {
@@ -120,6 +_,14 @@
 
     public BlockPos offset(int x, int y, int z) {
         return x == 0 && y == 0 && z == 0 ? this : new BlockPos(this.getX() + x, this.getY() + y, this.getZ() + z);
+    }
+
+    public Vec3 getCenter() {
+        return Vec3.atCenterOf(this);
+    }
+
+    public Vec3 getBottomCenter() {
+        return Vec3.atBottomCenterOf(this);
     }
 
     public BlockPos offset(Vec3i vec) {
