--- a/net/minecraft/core/HolderLookup.java
+++ b/net/minecraft/core/HolderLookup.java
@@ -69,6 +_,8 @@
     }
 
     interface RegistryLookup<T> extends HolderLookup<T>, HolderOwner<T> {
+        Optional<T> getValueForCopying(ResourceKey<T> resourceKey); // Paper - add method to get the value for pre-filling builders in the reg mod API
+
         ResourceKey<? extends Registry<? extends T>> key();
 
         Lifecycle registryLifecycle();
@@ -79,6 +_,12 @@
 
         default HolderLookup.RegistryLookup<T> filterElements(Predicate<T> filter) {
             return new HolderLookup.RegistryLookup.Delegate<T>() {
+                // Paper start - add getValueForCopying
+                @Override
+                public Optional<T> getValueForCopying(final ResourceKey<T> resourceKey) {
+                    return this.parent().getValueForCopying(resourceKey).filter(filter);
+                }
+                // Paper end - add getValueForCopying
                 @Override
                 public HolderLookup.RegistryLookup<T> parent() {
                     return RegistryLookup.this;
@@ -96,9 +_,34 @@
             };
         }
 
+        /// {@return the data map value attached with the object with the key, or {@code null} if there's no attached value}
+        ///
+        /// @param type the type of the data map
+        /// @param key  the object to get the value for
+        /// @param <A>  the data type
+        @org.jspecify.annotations.Nullable
+        default <A> A getData(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> type, ResourceKey<T> key) {
+            return null;
+        }
+
+        /// {@return the data map of the given `type`}
+        ///
+        /// @param type the type of the data map
+        /// @param <A>  the data type
+        default <A> Map<ResourceKey<T>, A> getDataMap(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> type) {
+            return Map.of();
+        }
+
         interface Delegate<T> extends HolderLookup.RegistryLookup<T> {
             HolderLookup.RegistryLookup<T> parent();
 
+            // Paper start - add getValueForCopying
+            @Override
+            default Optional<T> getValueForCopying(ResourceKey<T> resourceKey) {
+                return this.parent().getValueForCopying(resourceKey);
+            }
+            // Paper end - add getValueForCopying
+
             @Override
             default ResourceKey<? extends Registry<? extends T>> key() {
                 return this.parent().key();
@@ -127,6 +_,17 @@
             @Override
             default Stream<HolderSet.Named<T>> listTags() {
                 return this.parent().listTags();
+            }
+
+            @Override
+            @org.jspecify.annotations.Nullable
+            default <A> A getData(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> type, ResourceKey<T> key) {
+                return parent().getData(type, key);
+            }
+
+            @Override
+            default <A> Map<ResourceKey<T>, A> getDataMap(net.neoforged.neoforge.registries.datamaps.DataMapType<T, A> type) {
+                return parent().getDataMap(type);
             }
         }
     }
