--- a/net/minecraft/core/RegistrySynchronization.java
+++ b/net/minecraft/core/RegistrySynchronization.java
@@ -48,6 +_,9 @@
                     registry.listElements()
                         .forEach(
                             element -> {
+                                if (element.key().identifier().getNamespace().equals("paper")) {
+                                    return;
+                                }
                                 boolean canSkipContents = registry.registrationInfo(element.key())
                                     .flatMap(RegistrationInfo::knownPackInfo)
                                     .filter(clientKnownPacks::contains)
