--- a/net/minecraft/core/Rotations.java
+++ b/net/minecraft/core/Rotations.java
@@ -32,6 +_,18 @@
         this(p_123154_.getFloat(0), p_123154_.getFloat(1), p_123154_.getFloat(2));
     }
 
+    // Paper start - faster alternative constructor
+    private Rotations(float x, float y, float z, Void dummy_var) {
+        this.x = x;
+        this.y = y;
+        this.z = z;
+    }
+
+    public static Rotations createWithoutValidityChecks(float x, float y, float z) {
+        return new Rotations(x, y, z, null);
+    }
+    // Paper end - faster alternative constructor
+
     public ListTag save() {
         ListTag listtag = new ListTag();
         listtag.add(FloatTag.valueOf(this.x));
