--- a/net/minecraft/core/Vec3i.java
+++ b/net/minecraft/core/Vec3i.java
@@ -16,9 +_,9 @@
             p_123313_ -> IntStream.of(p_123313_.getX(), p_123313_.getY(), p_123313_.getZ())
         );
     public static final Vec3i ZERO = new Vec3i(0, 0, 0);
-    private int x;
+    public int x;
     private int y;
-    private int z;
+    public int z;
 
     public static Codec<Vec3i> offsetCodec(int p_194651_) {
         return CODEC.validate(
@@ -27,6 +_,12 @@
                     : DataResult.error(() -> "Position out of range, expected at most " + p_194651_ + ": " + p_274739_)
         );
     }
+
+    // Paper start
+    public final boolean isInsideBuildHeightAndWorldBoundsHorizontal(net.minecraft.world.level.LevelHeightAccessor levelHeightAccessor) {
+        return getX() >= -30000000 && getZ() >= -30000000 && getX() < 30000000 && getZ() < 30000000 && !levelHeightAccessor.isOutsideBuildHeight(getY());
+    }
+    // Paper end
 
     public Vec3i(int p_123296_, int p_123297_, int p_123298_) {
         this.x = p_123296_;
