--- a/net/minecraft/core/Vec3i.java
+++ b/net/minecraft/core/Vec3i.java
@@ -22,13 +_,13 @@
         ByteBufCodecs.VAR_INT, Vec3i::getX, ByteBufCodecs.VAR_INT, Vec3i::getY, ByteBufCodecs.VAR_INT, Vec3i::getZ, Vec3i::new
     );
     public static final Vec3i ZERO = new Vec3i(0, 0, 0);
-    private int x;
-    private int y;
-    private int z;
+    protected int x; // Paper - Perf: Manually inline methods in BlockPos; protected
+    protected int y; // Paper - Perf: Manually inline methods in BlockPos; protected
+    protected int z; // Paper - Perf: Manually inline methods in BlockPos; protected
 
     public static Codec<Vec3i> offsetCodec(int maxOffsetPerAxis) {
         return CODEC.validate(
-            value -> Math.abs(value.getX()) <= maxOffsetPerAxis && Math.abs(value.getY()) <= maxOffsetPerAxis && Math.abs(value.getZ()) <= maxOffsetPerAxis
+            value -> Math.abs(value.getX()) < maxOffsetPerAxis && Math.abs(value.getY()) < maxOffsetPerAxis && Math.abs(value.getZ()) < maxOffsetPerAxis
                 ? DataResult.success(value)
                 : DataResult.error(() -> "Position out of range, expected at most " + maxOffsetPerAxis + ": " + value)
         );
@@ -41,7 +_,7 @@
     }
 
     @Override
-    public boolean equals(Object o) {
+    public final boolean equals(Object o) { // Paper - Perf: Final for inline
         if (this == o) {
             return true;
         } else {
@@ -50,7 +_,7 @@
     }
 
     @Override
-    public int hashCode() {
+    public final int hashCode() { // Paper - Perf: Final for inline
         return (this.getY() + this.getZ() * 31) * 31 + this.getX();
     }
 
@@ -62,15 +_,15 @@
         }
     }
 
-    public int getX() {
+    public final int getX() { // Paper - Perf: Final for inline
         return this.x;
     }
 
-    public int getY() {
+    public final int getY() { // Paper - Perf: Final for inline
         return this.y;
     }
 
-    public int getZ() {
+    public final int getZ() { // Paper - Perf: Final for inline
         return this.z;
     }
 
@@ -234,10 +_,6 @@
         return Math.max(Math.max(xd, yd), zd);
     }
 
-    public boolean differsHorizontally(Vec3i pos) {
-        return this.x != pos.x || this.z != pos.z;
-    }
-
     public int get(Direction.Axis axis) {
         return axis.choose(this.x, this.y, this.z);
     }
@@ -254,4 +_,11 @@
     public String toShortString() {
         return this.getX() + ", " + this.getY() + ", " + this.getZ();
     }
+
+    // Paper start - Perf: Optimize isInWorldBounds
+    public final boolean isInsideBuildHeightAndWorldBoundsHorizontal(final net.minecraft.world.level.LevelHeightAccessor levelHeightAccessor) {
+        int maxSize = net.minecraft.world.level.Level.MAX_LEVEL_SIZE;
+        return this.getX() >= -maxSize && this.getZ() >= -maxSize && this.getX() < maxSize && this.getZ() < maxSize && !levelHeightAccessor.isOutsideBuildHeight(this.getY());
+    }
+    // Paper end - Perf: Optimize isInWorldBounds
 }
