--- a/net/minecraft/commands/arguments/coordinates/LocalCoordinates.java
+++ b/net/minecraft/commands/arguments/coordinates/LocalCoordinates.java
@@ -3,9 +_,10 @@
 import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import net.minecraft.commands.CommandSourceStack;
-import net.minecraft.util.Mth;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.joml.Quaterniond;
+import org.joml.Vector3d;
 
 public record LocalCoordinates(double left, double up, double forwards) implements Coordinates {
     public static final char PREFIX_LOCAL_COORDINATE = '^';
@@ -13,24 +_,13 @@
     @Override
     public Vec3 getPosition(CommandSourceStack sender) {
         Vec3 source = sender.getAnchor().apply(sender);
-        return this.apply(source, sender.getRotation());
+        Quaterniond localToWorld = localToWorld(sender.getRotation());
+        Vector3d relative = localToWorld.transform(this.left, this.up, this.forwards, new Vector3d());
+        return source.add(relative.x, relative.y, relative.z);
     }
 
-    public Vec3 apply(Vec3 source, Vec2 rotation) {
-        float yCos = Mth.cos((rotation.y + 90.0F) * (float) (Math.PI / 180.0));
-        float ySin = Mth.sin((rotation.y + 90.0F) * (float) (Math.PI / 180.0));
-        float xCos = Mth.cos(-rotation.x * (float) (Math.PI / 180.0));
-        float xSin = Mth.sin(-rotation.x * (float) (Math.PI / 180.0));
-        float xCosUp = Mth.cos((-rotation.x + 90.0F) * (float) (Math.PI / 180.0));
-        float xSinUp = Mth.sin((-rotation.x + 90.0F) * (float) (Math.PI / 180.0));
-        Vec3 forwards = new Vec3(yCos * xCos, xSin, ySin * xCos);
-        Vec3 up = new Vec3(yCos * xCosUp, xSinUp, ySin * xCosUp);
-        Vec3 left = forwards.cross(up).scale(-1.0);
-        return source.add(
-            forwards.x * this.forwards + up.x * this.up + left.x * this.left,
-            forwards.y * this.forwards + up.y * this.up + left.y * this.left,
-            forwards.z * this.forwards + up.z * this.up + left.z * this.left
-        );
+    private static Quaterniond localToWorld(Vec2 rotation) {
+        return new Quaterniond().rotationY(-rotation.y * (float) (Math.PI / 180.0)).rotateX(rotation.x * (float) (Math.PI / 180.0));
     }
 
     @Override
