--- a/net/minecraft/world/item/alchemy/PotionContents.java
+++ b/net/minecraft/world/item/alchemy/PotionContents.java
@@ -20,7 +_,6 @@
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.tags.TagKey;
 import net.minecraft.util.ARGB;
 import net.minecraft.util.Util;
 import net.minecraft.world.effect.MobEffect;
@@ -78,15 +_,7 @@
     }
 
     public boolean is(Holder<Potion> potion) {
-        return this.isPotionWithoutCustomEffects() && this.potion.get().is(potion);
-    }
-
-    public boolean is(TagKey<Potion> potion) {
-        return this.isPotionWithoutCustomEffects() && this.potion.get().is(potion);
-    }
-
-    public boolean isPotionWithoutCustomEffects() {
-        return this.potion.isPresent() && this.customEffects.isEmpty();
+        return this.potion.isPresent() && this.potion.get().is(potion) && this.customEffects.isEmpty();
     }
 
     public Iterable<MobEffectInstance> getAllEffects() {
@@ -167,7 +_,7 @@
                 if (effect.getEffect().value().isInstantaneous()) {
                     effect.getEffect().value().applyInstantaneousEffect(serverLevel, player, player, entity, effect.getAmplifier(), 1.0);
                 } else {
-                    entity.addEffect(effect);
+                    entity.addEffect(effect, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.POTION_DRINK);
                 }
             }, durationScale);
         }
@@ -197,6 +_,10 @@
         if (!modifiers.isEmpty()) {
             lines.accept(CommonComponents.EMPTY);
             lines.accept(Component.translatable("potion.whenDrank").withStyle(ChatFormatting.DARK_PURPLE));
+
+            // Neo: Override handling of potion attribute tooltips to support IAttributeExtension
+            net.neoforged.neoforge.common.util.AttributeUtil.addPotionTooltip(modifiers, lines);
+            if (true) return;
 
             for (Pair<Holder<Attribute>, AttributeModifier> entry : modifiers) {
                 AttributeModifier modifier = entry.getSecond();
