--- a/net/minecraft/client/gui/components/debug/DebugEntryCategory.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryCategory.java
@@ -1,16 +_,11 @@
 package net.minecraft.client.gui.components.debug;
 
-import net.minecraft.ChatFormatting;
 import net.minecraft.network.chat.Component;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
 public record DebugEntryCategory(Component label, float sortKey) {
-    public static final DebugEntryCategory SCREEN_TEXT = new DebugEntryCategory(
-        Component.translatable("debug.options.category.text").withStyle(ChatFormatting.BOLD, ChatFormatting.UNDERLINE), 1.0F
-    );
-    public static final DebugEntryCategory RENDERER = new DebugEntryCategory(
-        Component.translatable("debug.options.category.renderer").withStyle(ChatFormatting.BOLD, ChatFormatting.UNDERLINE), 2.0F
-    );
+    public static final DebugEntryCategory SCREEN_TEXT = new DebugEntryCategory(Component.translatable("debug.options.category.text"), 1.0F);
+    public static final DebugEntryCategory RENDERER = new DebugEntryCategory(Component.translatable("debug.options.category.renderer"), 2.0F);
 }
