--- a/net/minecraft/client/gui/components/debug/DebugEntryDetailedMemory.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryDetailedMemory.java
@@ -14,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class DebugEntryDetailedMemory implements DebugScreenEntry {
-    public static final Identifier GROUP = Identifier.withDefaultNamespace("memory");
+    private static final Identifier GROUP = Identifier.withDefaultNamespace("memory");
     private final MemoryMXBean memoryBean = ManagementFactory.getMemoryMXBean();
 
     @Override
