--- a/net/minecraft/client/gui/components/debug/DebugEntryChunkGeneration.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryChunkGeneration.java
@@ -14,7 +_,6 @@
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.levelgen.RandomState;
-import net.minecraft.world.level.levelgen.densityfunction.SamplerContext;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
@@ -44,11 +_,10 @@
         this.result.clear();
         this.lastPos = feetPos;
         ServerChunkCache chunkSource = serverLevel.getChunkSource();
-        SamplerContext samplerContext = SamplerContext.builder().enableCaches().build();
         ChunkGenerator generator = chunkSource.getGenerator();
         RandomState randomState = chunkSource.randomState();
-        generator.addDebugScreenInfo(this.result, randomState, feetPos, samplerContext);
-        Climate.Sampler sampler = randomState.createClimateSampler(samplerContext);
+        generator.addDebugScreenInfo(this.result, randomState, feetPos);
+        Climate.Sampler sampler = randomState.sampler();
         BiomeSource biomeSource = generator.getBiomeSource();
         biomeSource.addDebugInfo(this.result, feetPos, sampler);
         if (serverChunk != null && serverChunk.isOldNoiseGeneration()) {
