--- a/net/minecraft/client/gui/components/debug/DebugEntryMemory.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryMemory.java
@@ -14,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class DebugEntryMemory implements DebugScreenEntry {
-    public static final Identifier GROUP = Identifier.withDefaultNamespace("memory");
+    private static final Identifier GROUP = Identifier.withDefaultNamespace("memory");
     private final DebugEntryMemory.AllocationRateCalculator allocationRateCalculator = new DebugEntryMemory.AllocationRateCalculator();
 
     @Override
