--- a/net/minecraft/client/gui/components/debug/DebugEntryHeightmap.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryHeightmap.java
@@ -33,7 +_,7 @@
             "ML"
         )
     );
-    public static final Identifier GROUP = Identifier.withDefaultNamespace("heightmaps");
+    private static final Identifier GROUP = Identifier.withDefaultNamespace("heightmaps");
 
     @Override
     public void display(DebugScreenDisplayer displayer, @Nullable Level serverOrClientLevel, @Nullable LevelChunk clientChunk, @Nullable LevelChunk serverChunk) {
