--- a/net/minecraft/client/gui/components/debug/DebugEntrySystemSpecs.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntrySystemSpecs.java
@@ -16,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class DebugEntrySystemSpecs implements DebugScreenEntry {
-    public static final Identifier GROUP = Identifier.withDefaultNamespace("system");
+    private static final Identifier GROUP = Identifier.withDefaultNamespace("system");
 
     @Override
     public void display(DebugScreenDisplayer displayer, @Nullable Level serverOrClientLevel, @Nullable LevelChunk clientChunk, @Nullable LevelChunk serverChunk) {
