--- a/net/minecraft/client/gui/components/debug/DebugEntryBiome.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryBiome.java
@@ -17,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class DebugEntryBiome implements DebugScreenEntry {
-    public static final Identifier GROUP = Identifier.withDefaultNamespace("biome");
+    private static final Identifier GROUP = Identifier.withDefaultNamespace("biome");
 
     @Override
     public void display(DebugScreenDisplayer displayer, @Nullable Level serverOrClientLevel, @Nullable LevelChunk clientChunk, @Nullable LevelChunk serverChunk) {
