--- a/net/minecraft/client/gui/components/debug/DebugEntryChunkGeneration.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryChunkGeneration.java
@@ -20,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class DebugEntryChunkGeneration implements DebugScreenEntry {
-    public static final Identifier GROUP = Identifier.withDefaultNamespace("chunk_generation");
+    private static final Identifier GROUP = Identifier.withDefaultNamespace("chunk_generation");
     private final List<String> result = new ArrayList<>();
     private @Nullable BlockPos lastPos = null;
 
