--- a/net/minecraft/client/gui/components/debug/DebugEntryLookingAt.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntryLookingAt.java
@@ -26,8 +_,8 @@
 @OnlyIn(Dist.CLIENT)
 public abstract class DebugEntryLookingAt implements DebugScreenEntry {
     private static final int RANGE = 20;
-    public static final Identifier BLOCK_GROUP = Identifier.withDefaultNamespace("looking_at_block");
-    public static final Identifier FLUID_GROUP = Identifier.withDefaultNamespace("looking_at_fluid");
+    private static final Identifier BLOCK_GROUP = Identifier.withDefaultNamespace("looking_at_block");
+    private static final Identifier FLUID_GROUP = Identifier.withDefaultNamespace("looking_at_fluid");
 
     @Override
     public void display(DebugScreenDisplayer displayer, @Nullable Level serverOrClientLevel, @Nullable LevelChunk clientChunk, @Nullable LevelChunk serverChunk) {
