--- a/net/minecraft/client/gui/components/debug/DebugEntrySimplePerformanceImpactors.java
+++ b/net/minecraft/client/gui/components/debug/DebugEntrySimplePerformanceImpactors.java
@@ -32,9 +_,6 @@
         } else {
             displayer.addLine(String.format(Locale.ROOT, "Filtering: %s", filteringMethod.caption().getString()));
         }
-
-        boolean isMultiDrawIndirect = minecraft.levelRenderer.isChunkRenderingUsingMultiDrawIndirect();
-        displayer.addLine(String.format(Locale.ROOT, "Terrain Rendering: %s", isMultiDrawIndirect ? "multidrawindirect" : "naive"));
     }
 
     @Override
