--- a/net/minecraft/server/jsonrpc/Connection.java
+++ b/net/minecraft/server/jsonrpc/Connection.java
@@ -98,6 +_,7 @@
             super.exceptionCaught(ctx, cause);
             this.channel.close().awaitUninterruptibly();
         }
+        if (net.minecraft.server.MinecraftServer.getServer().isDebugging()) cause.printStackTrace(); // Spigot
     }
 
     protected void channelRead0(ChannelHandlerContext channelHandlerContext, JsonElement jsonElement) {
@@ -206,10 +_,10 @@
             JsonElement result = this.dispatchIncomingRequest(method, params);
             return result != null && sendResponse ? JsonRPCUtils.createSuccessResult(id, result) : null;
         } catch (InvalidParameterJsonRpcException e) {
-            LOGGER.debug("Invalid parameter invocation {}: {}", method, params, e);
+            LOGGER.debug("Invalid parameter invocation {}: {}, {}", method, params, e.getMessage());
             return sendResponse ? JsonRPCErrors.INVALID_PARAMS.create(id, e.getMessage()) : null;
         } catch (EncodeJsonRpcException e) {
-            LOGGER.error("Failed to encode json rpc response {}", method, e);
+            LOGGER.error("Failed to encode json rpc response {}: {}", method, e.getMessage());
             return sendResponse ? JsonRPCErrors.INTERNAL_ERROR.create(id, e.getMessage()) : null;
         } catch (InvalidRequestJsonRpcException e) {
             return sendResponse ? JsonRPCErrors.INVALID_REQUEST.create(id, e.getMessage()) : null;
