--- a/net/minecraft/server/jsonrpc/Connection.java
+++ b/net/minecraft/server/jsonrpc/Connection.java
@@ -98,6 +_,7 @@
             super.exceptionCaught(ctx, cause);
             this.channel.close().awaitUninterruptibly();
         }
+        if (net.minecraft.server.MinecraftServer.getServer().isDebugging()) cause.printStackTrace(); // Spigot
     }
 
     protected void channelRead0(ChannelHandlerContext channelHandlerContext, JsonElement jsonElement) {
