--- a/net/minecraft/server/jsonrpc/JsonRpcNotificationService.java
+++ b/net/minecraft/server/jsonrpc/JsonRpcNotificationService.java
@@ -61,26 +_,6 @@
     }
 
     @Override
-    public void worldUpgradeStarted() {
-        this.broadcastNotification(OutgoingRpcMethods.WORLD_UPGRADE_STARTED);
-    }
-
-    @Override
-    public void worldUpgradeProgress(float progressPercentage) {
-        this.broadcastNotification(OutgoingRpcMethods.WORLD_UPGRADE_PROGRESS, progressPercentage);
-    }
-
-    @Override
-    public void worldUpgradeFinished() {
-        this.broadcastNotification(OutgoingRpcMethods.WORLD_UPGRADE_FINISHED);
-    }
-
-    @Override
-    public void worldUpgradeFailed(String reason) {
-        this.broadcastNotification(OutgoingRpcMethods.WORLD_UPGRADE_FAILED, reason);
-    }
-
-    @Override
     public void playerOped(ServerOpListEntry operator) {
         this.broadcastNotification(OutgoingRpcMethods.PLAYER_OPED, OperatorService.OperatorDto.from(operator));
     }
@@ -121,7 +_,8 @@
     }
 
     @Override
-    public <T> void onGameRuleChanged(GameRule<T> gameRule, T value) {
+    public <T> void onGameRuleChanged(net.minecraft.server.level.ServerLevel level, GameRule<T> gameRule, T value) { // Paper - per-world game rules (add level param)
+        if (level != level.getServer().overworld()) return; // Paper - per-world game rules - use overworld for vanilla protocol
         this.broadcastNotification(OutgoingRpcMethods.GAMERULE_CHANGED, GameRulesService.getTypedRule(this.minecraftApi, gameRule, value));
     }
 
