--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -24,7 +_,7 @@
         AbstractPiglin.class, EntityDataSerializers.BOOLEAN
     );
     protected static final int CONVERSION_TIME = 300;
-    public int timeInOverworld;
+    protected int timeInOverworld;
 
     public AbstractPiglin(EntityType<? extends AbstractPiglin> p_34652_, Level p_34653_) {
         super(p_34652_, p_34653_);
@@ -46,7 +_,7 @@
         this.getEntityData().set(DATA_IMMUNE_TO_ZOMBIFICATION, p_34671_);
     }
 
-    public boolean isImmuneToZombification() {
+    protected boolean isImmuneToZombification() {
         return this.getEntityData().get(DATA_IMMUNE_TO_ZOMBIFICATION);
     }
 
@@ -82,7 +_,7 @@
             this.timeInOverworld = 0;
         }
 
-        if (this.timeInOverworld > 300) {
+        if (this.timeInOverworld > 300 && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.ZOMBIFIED_PIGLIN, (timer) -> this.timeInOverworld = timer)) {
             this.playConvertedSound();
             this.finishConversion((ServerLevel)this.level());
         }
@@ -96,6 +_,7 @@
         ZombifiedPiglin zombifiedpiglin = this.convertTo(EntityType.ZOMBIFIED_PIGLIN, true);
         if (zombifiedpiglin != null) {
             zombifiedpiglin.addEffect(new MobEffectInstance(MobEffects.CONFUSION, 200, 0));
+            net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, zombifiedpiglin);
         }
     }
 
