--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -1,8 +_,13 @@
 package net.minecraft.world.entity.monster.piglin;
 
+import com.mojang.serialization.Codec;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.component.DataComponents;
+import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
@@ -97,18 +_,74 @@
         List.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.NEAREST_PLAYERS, SensorType.NEAREST_ITEMS, SensorType.HURT_BY, SensorType.PIGLIN_SPECIFIC_SENSOR),
         PiglinAi::getActivities
     );
+    // CraftBukkit start - Custom bartering and interest list
+    public java.util.Set<Item> allowedBarterItems = new java.util.HashSet<>();
+    public java.util.Set<Item> interestItems = new java.util.HashSet<>();
+    private static final com.mojang.serialization.Codec<java.util.Set<net.minecraft.world.item.Item>> ITEM_SET_CODEC = net.minecraft.core.registries.BuiltInRegistries.ITEM
+            .byNameCodec().listOf().xmap(java.util.HashSet::new, List::copyOf);
+    // CraftBukkit end
 
     public Piglin(EntityType<? extends AbstractPiglin> type, Level level) {
         super(type, level);
         this.xpReward = 5;
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.piglinRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.piglinRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.piglinControllable;
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.piglinMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.piglinScale);
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.piglinTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.piglinAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
+    // Purpur start - Check mobGriefing Overrides
+    @Override
+    public Boolean checkEntityPickUpLootOverride() {
+        return this.level().purpurConfig.piglinCanPickUpLoot;
+    }
+    // Purpur end - Check mobGriefing Overrides
+
     @Override
     protected void addAdditionalSaveData(ValueOutput output) {
         super.addAdditionalSaveData(output);
         output.putBoolean("IsBaby", this.isBaby());
         output.putBoolean("CannotHunt", this.cannotHunt);
         this.writeInventoryToTag(output);
+        // CraftBukkit start
+        output.store("Bukkit.BarterList", ITEM_SET_CODEC, this.allowedBarterItems);
+        output.store("Bukkit.InterestList", ITEM_SET_CODEC, this.interestItems);
+        // CraftBukkit end
     }
 
     @Override
@@ -117,6 +_,10 @@
         this.setBaby(input.getBooleanOr("IsBaby", false));
         this.setCannotHunt(input.getBooleanOr("CannotHunt", false));
         this.readInventoryFromTag(input);
+        // CraftBukkit start
+        this.allowedBarterItems = input.read("Bukkit.BarterList", ITEM_SET_CODEC).orElseGet(java.util.HashSet::new);
+        this.interestItems = input.read("Bukkit.InterestList", ITEM_SET_CODEC).orElseGet(java.util.HashSet::new);
+        // CraftBukkit end
     }
 
     @VisibleForDebug
@@ -126,9 +_,10 @@
     }
 
     @Override
-    protected void dropCustomDeathLoot(ServerLevel level, DamageSource source, boolean killedByPlayer) {
+    protected void dropCustomDeathLoot(final ServerLevel level, final DamageSource source, final boolean killedByPlayer) {
         super.dropCustomDeathLoot(level, source, killedByPlayer);
-        this.inventory.removeAllItems().forEach(itemStack -> this.spawnAtLocation(level, itemStack));
+        this.inventory.getItems().stream().filter(item -> !item.isEmpty()).forEach(itemStack -> this.spawnAtLocation(level, itemStack)); // Paper - delay removal post event - taken from removeAllItems
+        this.postDeathEventTasks.add(this.inventory::clearContent); // Paper - only clear after death event
     }
 
     protected ItemStack addToInventory(ItemStack itemStack) {
@@ -166,6 +_,11 @@
     }
 
     public static boolean checkPiglinSpawnRules(EntityType<Piglin> type, LevelAccessor level, EntitySpawnReason spawnReason, BlockPos pos, RandomSource random) {
+        // Purpur start - Config to disable hostile mob spawn on ice
+        if (canSpawnInBlueAndPackedIce(level, pos)) {
+            return false;
+        }
+        // Purpur end - Config to disable hostile mob spawn on ice
         return !level.getBlockState(pos.below()).is(Blocks.NETHER_WART_BLOCK);
     }
 
@@ -321,7 +_,7 @@
         } else if (this.isChargingCrossbow()) {
             return PiglinArmPose.CROSSBOW_CHARGE;
         } else {
-            return this.isHolding(Items.CROSSBOW) && CrossbowItem.isCharged(this.getWeaponItem()) ? PiglinArmPose.CROSSBOW_HOLD : PiglinArmPose.DEFAULT;
+            return this.isHolding(is -> is.getItem() instanceof net.minecraft.world.item.CrossbowItem) && CrossbowItem.isCharged(this.getWeaponItem()) ? PiglinArmPose.CROSSBOW_HOLD : PiglinArmPose.DEFAULT;
         }
     }
 
@@ -350,7 +_,7 @@
 
     @Override
     public boolean canUseNonMeleeWeapon(ItemStack item) {
-        return item.getItem() == Items.CROSSBOW || item.has(DataComponents.KINETIC_WEAPON);
+        return item.is(net.neoforged.neoforge.common.Tags.Items.PIGLIN_USABLE_CROSSBOWS) || item.has(DataComponents.KINETIC_WEAPON);
     }
 
     protected void holdInMainHand(ItemStack itemStack) {
@@ -360,14 +_,14 @@
 
     protected void holdInOffHand(ItemStack itemStack) {
         this.setItemSlotAndDropWhenKilled(EquipmentSlot.OFFHAND, itemStack);
-        if (!itemStack.is(PiglinAi.BARTERING_ITEM)) {
+        if (!itemStack.isPiglinCurrency() && !allowedBarterItems.contains(itemStack.getItem())) { // CraftBukkit - Changes to accept custom payment items
             this.setPersistenceRequired();
         }
     }
 
     @Override
     public boolean wantsToPickUp(ServerLevel level, ItemStack itemStack) {
-        return level.getGameRules().get(GameRules.MOB_GRIEFING) && this.canPickUpLoot() && PiglinAi.wantsToPickup(this, itemStack);
+        return net.neoforged.neoforge.event.EventHooks.canEntityGrief(level, this) && this.canPickUpLoot() && PiglinAi.wantsToPickup(this, itemStack);
     }
 
     protected boolean canReplaceCurrentItem(ItemStack newItemStack) {
@@ -382,8 +_,8 @@
             return false;
         } else {
             TagKey<Item> preferredWeaponType = this.getPreferredWeaponType();
-            boolean newItemWanted = PiglinAi.isLovedItem(newItemStack) || preferredWeaponType != null && newItemStack.is(preferredWeaponType);
-            boolean currentItemWanted = PiglinAi.isLovedItem(currentItemStack) || preferredWeaponType != null && currentItemStack.is(preferredWeaponType);
+            boolean newItemWanted = PiglinAi.isLovedItem(newItemStack, this) || preferredWeaponType != null && newItemStack.is(preferredWeaponType); // CraftBukkit
+            boolean currentItemWanted = PiglinAi.isLovedItem(currentItemStack, this) || preferredWeaponType != null && currentItemStack.is(preferredWeaponType); // CraftBukkit
             if (newItemWanted && !currentItemWanted) {
                 return true;
             } else {
@@ -394,7 +_,7 @@
 
     @Override
     protected void pickUpItem(ServerLevel level, ItemEntity entity) {
-        this.onItemPickup(entity);
+        // this.onItemPickup(entity); // Paper - EntityPickupItemEvent fixes; call in PiglinAi#pickUpItem after EntityPickupItemEvent is fired
         PiglinAi.pickUpItem(level, this, entity);
     }
 
