--- a/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
+++ b/net/minecraft/world/entity/monster/piglin/PiglinBrute.java
@@ -140,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.PIGLIN_BRUTE_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_35072_) {
+    public SoundEvent getHurtSound(DamageSource p_35072_) {
         return SoundEvents.PIGLIN_BRUTE_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.PIGLIN_BRUTE_DEATH;
     }
 
