--- a/net/minecraft/client/renderer/debug/SupportBlockRenderer.java
+++ b/net/minecraft/client/renderer/debug/SupportBlockRenderer.java
@@ -45,7 +_,7 @@
         }
 
         for (Entity entity : this.surroundEntities) {
-            if (entity != player && !entity.isInvisible()) {
+            if (entity != player) {
                 this.drawHighlights(entity, () -> this.getBias(entity), -16711936);
             }
         }
