--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -15,6 +_,7 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
@@ -68,18 +_,19 @@
     public static final int TIME_BETWEEN_PLAYER_SCANS = 20;
     private static final int ARENA_SIZE_CHUNKS = 8;
     public static final int ARENA_TICKET_LEVEL = 9;
-    private static final int GATEWAY_COUNT = 20;
+    public static final int GATEWAY_COUNT = 20;
     private static final int GATEWAY_DISTANCE = 96;
     public static final int DRAGON_SPAWN_Y = 128;
     private final Predicate<Entity> validPlayer;
+    private static final Component DEFAULT_BOSS_EVENT_NAME = Component.translatable("entity.minecraft.ender_dragon"); // Paper - ensure reset EnderDragon boss event name
     public final ServerBossEvent dragonEvent = (ServerBossEvent)new ServerBossEvent(
-            Component.translatable("entity.minecraft.ender_dragon"), BossEvent.BossBarColor.PINK, BossEvent.BossBarOverlay.PROGRESS
+            DEFAULT_BOSS_EVENT_NAME, BossEvent.BossBarColor.PINK, BossEvent.BossBarOverlay.PROGRESS
         )
         .setPlayBossMusic(true)
         .setCreateWorldFog(true);
     public final ServerLevel level;
     private final BlockPos origin;
-    private final ObjectArrayList<Integer> gateways = new ObjectArrayList<>();
+    public final ObjectArrayList<Integer> gateways = new ObjectArrayList<>();
     private final BlockPattern exitPortalPattern;
     private int ticksSinceDragonSeen;
     private int crystalsAlive;
@@ -97,7 +_,7 @@
     public DragonRespawnAnimation respawnStage;
     private int respawnTime;
     @Nullable
-    private List<EndCrystal> respawnCrystals;
+    public List<EndCrystal> respawnCrystals;
 
     public EndDragonFight(ServerLevel p_289759_, long p_289805_, EndDragonFight.Data p_289800_) {
         this(p_289759_, p_289805_, p_289800_, BlockPos.ZERO);
@@ -115,6 +_,12 @@
         if (p_289768_.isRespawning) {
             this.respawnStage = DragonRespawnAnimation.START;
         }
+        // Paper start - Add config to disable ender dragon legacy check
+        if (p_289768_ == EndDragonFight.Data.DEFAULT && !p_289771_.paperConfig().entities.spawning.scanForLegacyEnderDragon) {
+            this.needsStateScanning = false;
+            this.dragonKilled = true;
+        }
+        // Paper end - Add config to disable ender dragon legacy check
 
         this.portalLocation = p_289768_.exitPortalLocation.orElse(null);
         this.gateways.addAll(p_289768_.gateways.orElseGet(() -> {
@@ -212,9 +_,9 @@
             this.dragonUUID = enderdragon.getUUID();
             LOGGER.info("Found that there's a dragon still alive ({})", enderdragon);
             this.dragonKilled = false;
-            if (!flag) {
+            if (!flag && this.level.paperConfig().entities.behavior.shouldRemoveDragon) { // Paper - Toggle for removing existing dragon
                 LOGGER.info("But we didn't have a portal, let's remove it.");
-                enderdragon.discard();
+                enderdragon.discard(null); // CraftBukkit - add Bukkit remove cause
                 this.dragonUUID = null;
             }
         }
@@ -370,12 +_,22 @@
             this.dragonEvent.setVisible(false);
             this.spawnExitPortal(true);
             this.spawnNewGateway();
-            if (!this.previouslyKilled) {
-                this.level
-                    .setBlockAndUpdate(
-                        this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.getLocation(this.origin)),
-                        Blocks.DRAGON_EGG.defaultBlockState()
-                    );
+            // Paper start - Add DragonEggFormEvent
+            BlockPos eggPosition = this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.getLocation(this.origin));
+            org.bukkit.craftbukkit.block.CraftBlockState eggState = org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(this.level, eggPosition);
+            eggState.setData(Blocks.DRAGON_EGG.defaultBlockState());
+            io.papermc.paper.event.block.DragonEggFormEvent eggEvent = new io.papermc.paper.event.block.DragonEggFormEvent(org.bukkit.craftbukkit.block.CraftBlock.at(this.level, eggPosition), eggState,
+                    new org.bukkit.craftbukkit.boss.CraftDragonBattle(this));
+            // Paper end - Add DragonEggFormEvent
+            if (this.level.paperConfig().entities.behavior.enderDragonsDeathAlwaysPlacesDragonEgg || !this.previouslyKilled) { // Paper - Add toggle for always placing the dragon egg
+                // Paper start - Add DragonEggFormEvent
+                // this.level.setBlockAndUpdate(this.level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING, EndPodiumFeature.getLocation(this.origin)), Blocks.DRAGON_EGG.defaultBlockState());
+            } else {
+                eggEvent.setCancelled(true);
+            }
+            if (eggEvent.callEvent()) {
+                eggEvent.getNewState().update(true);
+                // Paper end - Add DragonEggFormEvent
             }
 
             this.previouslyKilled = true;
@@ -389,6 +_,24 @@
         this.gateways.clear();
     }
 
+    // Paper start - More DragonBattle API
+    public boolean spawnNewGatewayIfPossible() {
+        if (!this.gateways.isEmpty()) {
+            this.spawnNewGateway();
+            return true;
+        }
+        return false;
+    }
+
+    public List<EndCrystal> getSpikeCrystals() {
+        final List<EndCrystal> endCrystals = new java.util.ArrayList<>();
+        for (final SpikeFeature.EndSpike spike : SpikeFeature.getSpikesForLevel(this.level)) {
+            endCrystals.addAll(this.level.getEntitiesOfClass(EndCrystal.class, spike.getTopBoundingBox()));
+        }
+        return endCrystals;
+    }
+    // Paper end - More DragonBattle API
+
     private void spawnNewGateway() {
         if (!this.gateways.isEmpty()) {
             int i = this.gateways.remove(this.gateways.size() - 1);
@@ -398,7 +_,7 @@
         }
     }
 
-    private void spawnNewGateway(BlockPos p_64090_) {
+    public void spawnNewGateway(BlockPos p_64090_) {
         this.level.levelEvent(3000, p_64090_, 0);
         this.level
             .registryAccess()
@@ -416,7 +_,11 @@
                 this.portalLocation = this.portalLocation.below();
             }
         }
-
+        // Paper start - Prevent "softlocked" exit portal generation
+        if (this.portalLocation.getY() <= this.level.getMinBuildHeight()) {
+            this.portalLocation = this.portalLocation.atY(this.level.getMinBuildHeight() + 1);
+        }
+        // Paper end - Prevent "softlocked" exit portal generation
         if (endpodiumfeature.place(
             FeatureConfiguration.NONE, this.level, this.level.getChunkSource().getGenerator(), RandomSource.create(), this.portalLocation
         )) {
@@ -438,6 +_,7 @@
             );
             this.level.addFreshEntity(enderdragon);
             this.dragonUUID = enderdragon.getUUID();
+            this.resetSpikeCrystals(); // Paper - Reset ender crystals on dragon spawn
         }
 
         return enderdragon;
@@ -449,6 +_,10 @@
             this.ticksSinceDragonSeen = 0;
             if (p_64097_.hasCustomName()) {
                 this.dragonEvent.setName(p_64097_.getDisplayName());
+                // Paper start - ensure reset EnderDragon boss event name
+            } else {
+                this.dragonEvent.setName(DEFAULT_BOSS_EVENT_NAME);
+                // Paper end - ensure reset EnderDragon boss event name
             }
         }
     }
@@ -477,6 +_,10 @@
         return this.previouslyKilled;
     }
 
+    private AtomicBoolean tryRespawnBoolean = new AtomicBoolean(false);
+    public boolean tryRespawnBoolean(){
+        return tryRespawnBoolean.getAndSet(false);
+    }
     public void tryRespawn() {
         if (this.dragonKilled && this.respawnStage == null) {
             BlockPos blockpos = this.portalLocation;
@@ -499,6 +_,7 @@
             for (Direction direction : Direction.Plane.HORIZONTAL) {
                 List<EndCrystal> list = this.level.getEntitiesOfClass(EndCrystal.class, new AABB(blockpos1.relative(direction, 2)));
                 if (list.isEmpty()) {
+                    tryRespawnBoolean.set(false);
                     return;
                 }
 
@@ -507,7 +_,14 @@
 
             LOGGER.debug("Found all crystals, respawning dragon.");
             this.respawnDragon(list1);
+            tryRespawnBoolean.set(respawnDragonBoolean());
         }
+        tryRespawnBoolean.set(false);
+    }
+
+    private AtomicBoolean respawnDragonBoolean = new AtomicBoolean(false);
+    public boolean respawnDragonBoolean(){
+        return tryRespawnBoolean.getAndSet(false);
     }
 
     public void respawnDragon(List<EndCrystal> p_64092_) {
@@ -532,7 +_,9 @@
             this.respawnTime = 0;
             this.spawnExitPortal(false);
             this.respawnCrystals = p_64092_;
+            respawnDragonBoolean.set(true);
         }
+        respawnDragonBoolean.set(false);
     }
 
     public void resetSpikeCrystals() {
@@ -542,6 +_,14 @@
                 endcrystal.setBeamTarget(null);
             }
         }
+    }
+
+    public void addPlayer(ServerPlayer player) {
+        this.dragonEvent.addPlayer(player);
+    }
+
+    public void removePlayer(ServerPlayer player) {
+        this.dragonEvent.removePlayer(player);
     }
 
     @Nullable
