--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -16,6 +_,7 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Predicate;
 import net.minecraft.advancements.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -516,6 +_,10 @@
         return this.hasPreviouslyKilledDragon;
     }
 
+    private final AtomicBoolean tryRespawnBoolean = new AtomicBoolean(false);
+    public boolean tryRespawnBoolean(){
+        return tryRespawnBoolean.getAndSet(false);
+    }
     public void tryRespawn() {
         if (this.dragonKilled && this.respawnStage == null) {
             BlockPos location = this.exitPortalLocation;
@@ -538,17 +_,24 @@
             for (Direction direction : Direction.Plane.HORIZONTAL) {
                 List<EndCrystal> found = this.level.getEntitiesOfClass(EndCrystal.class, new AABB(center.relative(direction, 3)));
                 if (found.isEmpty()) {
+                    tryRespawnBoolean.set(false);
                     return;
                 }
 
                 crystals.addAll(found);
+                tryRespawnBoolean.set(respawnDragonBoolean());
             }
 
             LOGGER.debug("Found all crystals, respawning dragon.");
             this.respawnDragon(crystals);
         }
+        tryRespawnBoolean.set(false);
     }
 
+    private AtomicBoolean respawnDragonBoolean = new AtomicBoolean(false);
+    public boolean respawnDragonBoolean(){
+        return respawnDragonBoolean.getAndSet(false);
+    }
     public void respawnDragon(List<EndCrystal> crystals) {
         if (this.dragonKilled && this.respawnStage == null) {
             for (BlockPattern.BlockPatternMatch portal = this.findExitPortal(); portal != null; portal = this.findExitPortal()) {
@@ -569,7 +_,9 @@
             this.spawnExitPortal(false);
             this.respawnCrystals = crystals.stream().map(EntityReference::of).toList();
             this.setDirty();
+            respawnDragonBoolean.set(true);
         }
+        respawnDragonBoolean.set(false);
     }
 
     public void resetSpikeCrystals() {
@@ -579,6 +_,14 @@
                 crystal.setBeamTarget(null);
             }
         }
+    }
+
+    public void addPlayer(ServerPlayer player) {
+        this.dragonEvent.addPlayer(player);
+    }
+
+    public void removePlayer(ServerPlayer player) {
+        this.dragonEvent.removePlayer(player);
     }
 
     public @Nullable UUID dragonUUID() {
