--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -16,6 +_,7 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Predicate;
 import net.minecraft.advancements.triggers.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
@@ -423,14 +_,26 @@
         this.setDirty();
     }
 
-    private void spawnNewGateway() {
+    // Paper start - More DragonBattle API
+    public List<EndCrystal> getSpikeCrystals() {
+        final List<EndCrystal> crystals = new ArrayList<>();
+        for (final EndSpikeFeature.EndSpike spike : EndSpikeFeature.getSpikesForLevel(this.level)) {
+            crystals.addAll(this.level.getEntitiesOfClass(EndCrystal.class, spike.getTopBoundingBox()));
+        }
+        return crystals;
+    }
+    // Paper end - More DragonBattle API
+
+    public boolean spawnNewGateway() { // Paper - More DragonBattle API
         if (!this.gateways.isEmpty()) {
             int gateway = this.gateways.remove(this.gateways.size() - 1);
             int x = Mth.floor(96.0 * Math.cos(2.0 * (-Math.PI + (Math.PI / 20) * gateway)));
             int z = Mth.floor(96.0 * Math.sin(2.0 * (-Math.PI + (Math.PI / 20) * gateway)));
             this.spawnNewGateway(new BlockPos(x, 75, z));
             this.setDirty();
+            return true; // Paper - More DragonBattle API
         }
+        return false; // Paper - More DragonBattle API
     }
 
     public void spawnNewGateway(BlockPos pos) {
@@ -515,6 +_,10 @@
         return this.hasPreviouslyKilledDragon;
     }
 
+    private final AtomicBoolean tryRespawnBoolean = new AtomicBoolean(false);
+    public boolean tryRespawnBoolean(){
+        return tryRespawnBoolean.getAndSet(false);
+    }
     public void tryRespawn() {
         if (this.dragonKilled && this.respawnStage == null) {
             BlockPos location = this.exitPortalLocation;
@@ -537,17 +_,24 @@
             for (Direction direction : Direction.Plane.HORIZONTAL) {
                 List<EndCrystal> found = this.level.getEntitiesOfClass(EndCrystal.class, new AABB(center.relative(direction, 3)));
                 if (found.isEmpty()) {
+                    tryRespawnBoolean.set(false);
                     return;
                 }
 
                 crystals.addAll(found);
+                tryRespawnBoolean.set(respawnDragonBoolean());
             }
 
             LOGGER.debug("Found all crystals, respawning dragon.");
             this.respawnDragon(crystals);
         }
+        tryRespawnBoolean.set(false);
     }
 
+    private AtomicBoolean respawnDragonBoolean = new AtomicBoolean(false);
+    public boolean respawnDragonBoolean(){
+        return respawnDragonBoolean.getAndSet(false);
+    }
     public void respawnDragon(List<EndCrystal> crystals) {
         if (this.dragonKilled && this.respawnStage == null) {
             for (BlockPattern.BlockPatternMatch portal = this.findExitPortal(); portal != null; portal = this.findExitPortal()) {
@@ -568,7 +_,9 @@
             this.spawnExitPortal(false);
             this.respawnCrystals = crystals.stream().map(EntityReference::of).toList();
             this.setDirty();
+            respawnDragonBoolean.set(true);
         }
+        respawnDragonBoolean.set(false);
     }
 
     public void resetSpikeCrystals() {
@@ -578,6 +_,14 @@
                 crystal.setBeamTarget(null);
             }
         }
+    }
+
+    public void addPlayer(ServerPlayer player) {
+        this.dragonEvent.addPlayer(player);
+    }
+
+    public void removePlayer(ServerPlayer player) {
+        this.dragonEvent.removePlayer(player);
     }
 
     public @Nullable UUID dragonUUID() {
