--- a/net/minecraft/world/entity/ConversionType.java
+++ b/net/minecraft/world/entity/ConversionType.java
@@ -13,6 +13,11 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.scores.Scoreboard;
 
+// CraftBukkit start
+import net.minecraft.core.BlockPos;
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public enum ConversionType {
 
     SINGLE(true) {
@@ -28,7 +33,7 @@
 
                 for (Entity entity1 : mob1.getPassengers()) {
                     entity1.stopRiding();
-                    entity1.remove(Entity.RemovalReason.DISCARDED);
+                    entity1.remove(Entity.RemovalReason.DISCARDED, EntityRemoveEvent.Cause.TRANSFORMATION); // CraftBukkit - add Bukkit remove cause
                 }
 
                 entity.startRiding(mob1);
@@ -46,7 +51,7 @@
                     ItemStack itemstack = mob.getItemBySlot(equipmentslot);
 
                     if (!itemstack.isEmpty()) {
-                        mob1.setItemSlot(equipmentslot, itemstack.copyAndClear());
+                        mob1.setItemSlot(equipmentslot, itemstack.copy()); // CraftBukkit - SPIGOT-7996: don't clear yet
                         mob1.setDropChance(equipmentslot, mob.getDropChances().byEquipment(equipmentslot));
                     }
                 }
@@ -58,7 +63,7 @@
             mob1.hurtTime = mob.hurtTime;
             mob1.yBodyRot = mob.yBodyRot;
             mob1.setOnGround(mob.onGround());
-            Optional optional = mob.getSleepingPos();
+            Optional<BlockPos> optional = mob.getSleepingPos(); // CraftBukkit - decompile error
 
             Objects.requireNonNull(mob1);
             optional.ifPresent(mob1::setSleepingPos);
@@ -70,6 +75,21 @@
 
             this.convertCommon(mob, mob1, conversionparams);
         }
+
+        // CraftBukkit start
+        @Override
+        void postConvert(Mob mob, Mob mob1, ConversionParams conversionparams) {
+            if (conversionparams.keepEquipment()) {
+                for (EquipmentSlot equipmentslot : EquipmentSlot.VALUES) {
+                    ItemStack itemstack = mob.getItemBySlot(equipmentslot);
+
+                    if (!itemstack.isEmpty()) {
+                        itemstack.setCount(0); // SPIGOT-7996: clear after conversion
+                    }
+                }
+            }
+        }
+        // CraftBukkit end
     },
     SPLIT_ON_DEATH(false) {
         @Override
@@ -103,6 +123,8 @@
 
     abstract void convert(Mob mob, Mob mob1, ConversionParams conversionparams);
 
+    void postConvert(Mob mob, Mob mob1, ConversionParams conversionparams) {} // CraftBukkit
+
     void convertCommon(Mob mob, Mob mob1, ConversionParams conversionparams) {
         mob1.setAbsorptionAmount(mob.getAbsorptionAmount());
 
@@ -145,7 +167,7 @@
         mob1.setNoGravity(mob.isNoGravity());
         mob1.setPortalCooldown(mob.getPortalCooldown());
         mob1.setSilent(mob.isSilent());
-        Set set = mob.getTags();
+        Set<String> set = mob.getTags(); // CraftBukkit - decompile error
 
         Objects.requireNonNull(mob1);
         set.forEach(mob1::addTag);
