--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -141,8 +_,10 @@
         return panel;
     }
 
+    private java.util.concurrent.CountDownLatch latch = new java.util.concurrent.CountDownLatch(1);
     public void start() {
         this.logAppenderThread.start();
+        latch.countDown();
     }
 
     public void close() {
@@ -155,7 +_,11 @@
         this.finalizers.forEach(Runnable::run);
     }
 
+    private static final java.util.regex.Pattern ANSI = java.util.regex.Pattern.compile("\\x1B\\[([0-9]{1,2}(;[0-9]{1,2})*)?[m|K]"); // CraftBukkit
     public void print(JTextArea console, JScrollPane scrollPane, String line) {
+        try {
+            latch.await();
+        } catch (InterruptedException e){} //Prevent logging until after constructor has ended.
         if (!SwingUtilities.isEventDispatchThread()) {
             SwingUtilities.invokeLater(() -> this.print(console, scrollPane, line));
         } else {
@@ -167,7 +_,7 @@
             }
 
             try {
-                document.insertString(document.getLength(), line, null);
+                document.insertString(document.getLength(), ANSI.matcher(line).replaceAll(""), null); // CraftBukkit
             } catch (BadLocationException var8) {
             }
 
