--- a/com/mojang/renderpearl/backend/util/TransientBlockAllocator.java
+++ b/com/mojang/renderpearl/backend/util/TransientBlockAllocator.java
@@ -1,6 +_,5 @@
 package com.mojang.renderpearl.backend.util;
 
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import it.unimi.dsi.fastutil.objects.ReferenceArrayList;
 import java.util.function.Consumer;
 import java.util.function.LongFunction;
@@ -9,10 +_,9 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jetbrains.annotations.Contract;
 import org.jspecify.annotations.Nullable;
-import org.lwjgl.system.MemoryUtil;
 
 @OnlyIn(Dist.CLIENT)
-public class TransientBlockAllocator<T extends TransientBlockAllocator.Allocator.Block> implements UncheckedAutoCloseable {
+public class TransientBlockAllocator<T> implements AutoCloseable {
     private final long blockSize;
     private final long maxAlignment;
     private final TransientBlockAllocator.Allocator<T> allocator;
@@ -58,16 +_,11 @@
         ReferenceArrayList<T> specialBlocksUsedThisRotation = this.specialBlocks.clone();
         this.specialBlocks.clear();
         return () -> {
-            this.freeBlocks.ensureCapacity(this.freeBlocks.size() + blocksUsedThisRotation.size());
-
-            for (T usedBlock : blocksUsedThisRotation) {
-                if (usedBlock.suboptimal()) {
-                    this.allocator.free(usedBlock);
-                } else {
-                    this.freeBlocks.add(usedBlock);
-                }
+            if (!blocksUsedThisRotation.isEmpty()) {
+                this.allocator.free(blocksUsedThisRotation.pop());
             }
 
+            this.freeBlocks.addAll(blocksUsedThisRotation);
             specialBlocksUsedThisRotation.forEach(this.allocator::free);
         };
     }
@@ -148,12 +_,12 @@
     public record Allocation<T>(T block, long offset, long size) {
     }
 
-    public interface Allocator<T extends TransientBlockAllocator.Allocator.Block> {
+    public interface Allocator<T> {
         T alloc(long size);
 
         void free(T t);
 
-        static <T extends TransientBlockAllocator.Allocator.Block> TransientBlockAllocator.Allocator<T> create(LongFunction<T> alloc, Consumer<T> free) {
+        static <T> TransientBlockAllocator.Allocator<T> create(LongFunction<T> alloc, Consumer<T> free) {
             return new TransientBlockAllocator.Allocator<T>() {
                 @Override
                 public T alloc(long size) {
@@ -165,29 +_,6 @@
                     free.accept(t);
                 }
             };
-        }
-
-        interface Block {
-            boolean suboptimal();
-        }
-
-        record CpuBlock(long address) implements TransientBlockAllocator.Allocator.Block {
-            public CpuBlock {
-                if (address == 0L) {
-                    throw new IllegalArgumentException("Failed to allocate CPU block");
-                }
-            }
-
-            @Override
-            public boolean suboptimal() {
-                return false;
-            }
-
-            public static TransientBlockAllocator.Allocator<TransientBlockAllocator.Allocator.CpuBlock> memalloc() {
-                return TransientBlockAllocator.Allocator.create(
-                    size -> new TransientBlockAllocator.Allocator.CpuBlock(MemoryUtil.nmemAlloc(size)), block -> MemoryUtil.nmemFree(block.address)
-                );
-            }
         }
     }
 }
