--- a/net/minecraft/util/context/ContextMap.java
+++ b/net/minecraft/util/context/ContextMap.java
@@ -1,31 +_,29 @@
 package net.minecraft.util.context;
 
 import com.google.common.collect.Sets;
-import it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap;
+import java.util.IdentityHashMap;
 import java.util.Map;
 import java.util.NoSuchElementException;
 import java.util.Set;
 import org.jetbrains.annotations.Contract;
 import org.jspecify.annotations.Nullable;
 
-public final class ContextMap {
-    public static final ContextMap EMPTY = new ContextMap(Map.of());
+public class ContextMap {
+    // Neo: Add EMPTY map for convenience.
+    public static final ContextMap EMPTY = new Builder().create(ContextKeySet.EMPTY);
+
     private final Map<ContextKey<?>, Object> params;
 
     private ContextMap(Map<ContextKey<?>, Object> params) {
         this.params = params;
     }
 
-    public static ContextMap.Builder builder() {
-        return new ContextMap.Builder();
-    }
-
     public boolean has(ContextKey<?> key) {
         return this.params.containsKey(key);
     }
 
     public <T> T getOrThrow(ContextKey<T> key) {
-        T value = this.get(key);
+        T value = (T)this.params.get(key);
         if (value == null) {
             throw new NoSuchElementException(key.name().toString());
         } else {
@@ -33,7 +_,7 @@
         }
     }
 
-    public <T> @Nullable T get(ContextKey<T> key) {
+    public <T> @Nullable T getOptional(ContextKey<T> key) {
         return (T)this.params.get(key);
     }
 
@@ -43,12 +_,14 @@
     }
 
     public static class Builder {
-        private final Map<ContextKey<?>, Object> params = new Reference2ObjectOpenHashMap<>();
+        private final Map<ContextKey<?>, Object> params = new IdentityHashMap<>();
 
-        private Builder() {
+        public <T> ContextMap.Builder withParameter(ContextKey<T> param, T value) {
+            this.params.put(param, value);
+            return this;
         }
 
-        public <T> ContextMap.Builder set(ContextKey<T> param, @Nullable T value) {
+        public <T> ContextMap.Builder withOptionalParameter(ContextKey<T> param, @Nullable T value) {
             if (value == null) {
                 this.params.remove(param);
             } else {
@@ -58,24 +_,29 @@
             return this;
         }
 
-        public <T> @Nullable T get(ContextKey<T> param) {
+        public <T> T getParameter(ContextKey<T> param) {
+            T value = (T)this.params.get(param);
+            if (value == null) {
+                throw new NoSuchElementException(param.name().toString());
+            } else {
+                return value;
+            }
+        }
+
+        public <T> @Nullable T getOptionalParameter(ContextKey<T> param) {
             return (T)this.params.get(param);
         }
 
-        public ContextMap build() {
-            return this.params.isEmpty() ? ContextMap.EMPTY : new ContextMap(new Reference2ObjectOpenHashMap<>(this.params));
-        }
-
-        public ContextMap buildAndValidate(ContextKeySet paramSet) {
+        public ContextMap create(ContextKeySet paramSet) {
             Set<ContextKey<?>> notAllowed = Sets.difference(this.params.keySet(), paramSet.allowed());
-            if (!notAllowed.isEmpty()) {
+            if (false && !notAllowed.isEmpty()) {
                 throw new IllegalArgumentException("Parameters not allowed in this parameter set: " + notAllowed);
             } else {
                 Set<ContextKey<?>> missingRequired = Sets.difference(paramSet.required(), this.params.keySet());
                 if (!missingRequired.isEmpty()) {
                     throw new IllegalArgumentException("Missing required parameters: " + missingRequired);
                 } else {
-                    return this.build();
+                    return new ContextMap(this.params);
                 }
             }
         }
