--- a/net/minecraft/world/scores/ScoreboardSaveData.java
+++ b/net/minecraft/world/scores/ScoreboardSaveData.java
@@ -149,6 +_,7 @@
         ListTag listtag = new ListTag();
 
         for (PlayerTeam playerteam : this.scoreboard.getPlayerTeams()) {
+            if (!io.papermc.paper.configuration.GlobalConfiguration.get().scoreboards.saveEmptyScoreboardTeams && playerteam.getPlayers().isEmpty()) continue; // Paper - Don't save empty scoreboard teams to scoreboard.dat
             CompoundTag compoundtag = new CompoundTag();
             compoundtag.putString("Name", playerteam.getName());
             compoundtag.putString("DisplayName", Component.Serializer.toJson(playerteam.getDisplayName(), p_331247_));
