--- a/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/DefendVillageTargetGoal.java
@@ -47,6 +_,7 @@
 
     @Override
     public void start() {
+        this.golem.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.DEFEND_VILLAGE, true); // CraftBukkit - reason
         this.golem.setTarget(this.potentialTarget);
         super.start();
     }
