--- a/net/minecraft/world/entity/ai/goal/target/OwnerHurtTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/OwnerHurtTargetGoal.java
@@ -37,6 +_,7 @@
 
     @Override
     public void start() {
+        this.mob.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit - reason
         this.mob.setTarget(this.ownerLastHurt);
         LivingEntity livingentity = this.tameAnimal.getOwner();
         if (livingentity != null) {
