--- a/net/minecraft/world/entity/ai/goal/target/OwnerHurtByTargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/OwnerHurtByTargetGoal.java
@@ -37,6 +_,7 @@
 
     @Override
     public void start() {
+        this.mob.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_OWNER, true); // CraftBukkit - reason
         this.mob.setTarget(this.ownerLastHurtBy);
         LivingEntity livingentity = this.tameAnimal.getOwner();
         if (livingentity != null) {
