--- a/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
+++ b/net/minecraft/world/entity/ai/goal/target/TargetGoal.java
@@ -63,7 +_,8 @@
                         }
                     }
 
-                    this.mob.setTarget(livingentity);
+                    this.mob.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit
+                    this.mob.setTarget(livingentity); // CraftBukkit
                     return true;
                 }
             }
@@ -83,7 +_,8 @@
 
     @Override
     public void stop() {
-        this.mob.setTarget(null);
+        this.mob.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit
+        this.mob.setTarget(null); // CraftBukkit
         this.targetMob = null;
     }
 
