--- a/net/minecraft/server/jsonrpc/internalapi/MinecraftGameRuleServiceImpl.java
+++ b/net/minecraft/server/jsonrpc/internalapi/MinecraftGameRuleServiceImpl.java
@@ -10,21 +_,27 @@
 
 public class MinecraftGameRuleServiceImpl implements MinecraftGameRuleService {
     private final DedicatedServer server;
-    private final GameRules gameRules;
+    // private final GameRules gameRules;  // CraftBukkit - per-world
     private final JsonRpcLogger jsonrpcLogger;
 
+    // CraftBukkit - per-world
+    private GameRules getGameRules() {
+        return this.server.getGameRules();
+    }
+    // CraftBukkit end
+
     public MinecraftGameRuleServiceImpl(DedicatedServer server, JsonRpcLogger jsonrpcLogger) {
         this.server = server;
-        this.gameRules = server.getGameRules();
+        // this.gameRules = server.getGameRules();  // CraftBukkit - per-world
         this.jsonrpcLogger = jsonrpcLogger;
     }
 
     @Override
     public <T> GameRulesService.GameRuleUpdate<T> updateGameRule(GameRulesService.GameRuleUpdate<T> update, ClientInfo clientInfo) {
         GameRule<T> gameRule = update.gameRule();
-        T oldValue = this.gameRules.get(gameRule);
+        T oldValue = this.getGameRules().get(gameRule); // CraftBukkit - per-world
         T newValue = update.value();
-        this.gameRules.set(gameRule, newValue, this.server);
+        this.getGameRules().set(gameRule, newValue, this.server.overworld()); // CraftBukkit - per-world
         this.jsonrpcLogger
             .log(clientInfo, "Game rule '{}' updated from '{}' to '{}'", gameRule.id(), gameRule.serialize(oldValue), gameRule.serialize(newValue));
         return update;
@@ -37,11 +_,11 @@
 
     @Override
     public Stream<GameRule<?>> getAvailableGameRules() {
-        return this.gameRules.availableRules();
+        return this.getGameRules().availableRules(); // CraftBukkit - per-world
     }
 
     @Override
     public <T> T getRuleValue(GameRule<T> gameRule) {
-        return this.gameRules.get(gameRule);
+        return (T) this.getGameRules().get(gameRule); // CraftBukkit - per-world
     }
 }
