--- a/net/minecraft/client/renderer/state/level/CameraRenderState.java
+++ b/net/minecraft/client/renderer/state/level/CameraRenderState.java
@@ -19,7 +_,6 @@
     public boolean initialized;
     public boolean isPanoramicMode;
     public boolean isFrustumCaptured;
-    public boolean isFirstPerson;
     public boolean smartCull;
     public Quaternionf orientation = new Quaternionf();
     public float cameraEntityPartialTicks;
