--- a/net/minecraft/client/renderer/state/level/LevelRenderState.java
+++ b/net/minecraft/client/renderer/state/level/LevelRenderState.java
@@ -11,7 +_,6 @@
 @OnlyIn(Dist.CLIENT)
 public class LevelRenderState {
     public CameraRenderState cameraRenderState = new CameraRenderState();
-    public final PlayerRenderState playerRenderState = new PlayerRenderState();
     public final List<SectionUpdateRenderState> sectionUpdateRenderStates = new ArrayList<>();
     public final List<EntityRenderState> entityRenderStates = new ArrayList<>();
     public final List<BlockEntityRenderState> blockEntityRenderStates = new ArrayList<>();
@@ -28,7 +_,6 @@
     public float cloudHeight;
     public boolean render3dCrosshair;
     public boolean renderWireframeTerrain;
-    public boolean shouldUseMultiDrawIndirectForTerrain;
     public @Nullable Runnable playerCompiledSectionCallback;
     public ChunkLoadingRenderState chunkLoadingRenderState = new ChunkLoadingRenderState();
     public boolean shouldResetChunkLayerSampler;
@@ -37,7 +_,6 @@
 
     public void reset() {
         this.sectionUpdateRenderStates.clear();
-        this.playerRenderState.reset();
         this.entityRenderStates.clear();
         this.blockEntityRenderStates.clear();
         this.blockBreakingRenderStates.clear();
