--- a/net/minecraft/client/renderer/state/level/SkyRenderState.java
+++ b/net/minecraft/client/renderer/state/level/SkyRenderState.java
@@ -4,8 +_,6 @@
 import net.minecraft.world.level.dimension.DimensionType;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
-import org.joml.Vector3fc;
-import org.joml.Vector4fc;
 
 @OnlyIn(Dist.CLIENT)
 public class SkyRenderState {
@@ -16,9 +_,9 @@
     public float starAngle;
     public float rainBrightness;
     public float starBrightness;
-    public Vector4fc sunriseAndSunsetColor;
+    public int sunriseAndSunsetColor;
     public MoonPhase moonPhase = MoonPhase.FULL_MOON;
-    public Vector3fc skyColor;
+    public int skyColor;
     public float endFlashIntensity;
     public float endFlashXAngle;
     public float endFlashYAngle;
