--- a/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
+++ b/net/minecraft/network/protocol/handshake/ClientIntentionPacket.java
@@ -16,7 +_,8 @@
     }
 
     private ClientIntentionPacket(FriendlyByteBuf input) {
-        this(input.readVarInt(), input.readUtf(255), input.readUnsignedShort(), ClientIntent.byId(input.readVarInt()));
+        // Spigot - increase max hostName length
+        this(input.readVarInt(), input.readUtf(Short.MAX_VALUE), input.readUnsignedShort(), ClientIntent.byId(input.readVarInt()));
     }
 
     private void write(FriendlyByteBuf output) {
