--- a/net/minecraft/world/level/levelgen/presets/WorldPreset.java
+++ b/net/minecraft/world/level/levelgen/presets/WorldPreset.java
@@ -10,7 +_,7 @@
 import java.util.Optional;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
+import net.minecraft.resources.RegistryFileCodec;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.levelgen.WorldDimensions;
@@ -21,7 +_,7 @@
                 .apply(i, WorldPreset::new)
         )
         .validate(WorldPreset::requireOverworld);
-    public static final Codec<Holder<WorldPreset>> CODEC = RegistryCodecs.holder(Registries.WORLD_PRESET, DIRECT_CODEC);
+    public static final Codec<Holder<WorldPreset>> CODEC = RegistryFileCodec.create(Registries.WORLD_PRESET, DIRECT_CODEC);
     private final Map<ResourceKey<LevelStem>, LevelStem> dimensions;
 
     public WorldPreset(Map<ResourceKey<LevelStem>, LevelStem> dimensions) {
