--- a/net/minecraft/world/entity/monster/illager/SpellcasterIllager.java
+++ b/net/minecraft/world/entity/monster/illager/SpellcasterIllager.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.monster.illager;
 
 import java.util.EnumSet;
+import java.util.Objects;
 import java.util.function.IntFunction;
 import net.minecraft.core.particles.ColorParticleOption;
 import net.minecraft.core.particles.ParticleTypes;
@@ -209,6 +_,11 @@
         public void tick() {
             this.attackWarmupDelay--;
             if (this.attackWarmupDelay == 0) {
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleEntitySpellCastEvent(SpellcasterIllager.this, this.getSpell())) {
+                    return;
+                }
+                // CraftBukkit end
                 this.performSpellCasting();
                 SpellcasterIllager.this.playSound(SpellcasterIllager.this.getCastingSoundEvent(), 1.0F, 1.0F);
             }
