--- a/net/minecraft/world/entity/monster/illager/Vindicator.java
+++ b/net/minecraft/world/entity/monster/illager/Vindicator.java
@@ -180,8 +_,8 @@
     }
 
     private static class VindicatorBreakDoorGoal extends BreakDoorGoal {
-        public VindicatorBreakDoorGoal(Mob mob) {
-            super(mob, 6, Vindicator.DOOR_BREAKING_PREDICATE);
+        public VindicatorBreakDoorGoal(final Mob mob) {
+            super(mob, 6, com.google.common.base.Predicates.in(mob.level().paperConfig().entities.behavior.doorBreakingDifficulty.getOrDefault(mob.getType(), mob.level().paperConfig().entities.behavior.doorBreakingDifficulty.get(net.minecraft.world.entity.EntityTypes.VINDICATOR)))); // Paper - Configurable door breaking difficulty
             this.setFlags(EnumSet.of(Goal.Flag.MOVE));
         }
 
