--- a/net/minecraft/world/level/levelgen/structure/BoundingBox.java
+++ b/net/minecraft/world/level/levelgen/structure/BoundingBox.java
@@ -101,11 +_,12 @@
     }
 
     public boolean intersects(BoundingBox other) {
-        return this.intersects(other.minX, other.minY, other.minZ, other.maxX, other.maxY, other.maxZ);
-    }
-
-    public boolean intersects(int minX, int minY, int minZ, int maxX, int maxY, int maxZ) {
-        return this.maxX >= minX && this.minX <= maxX && this.maxZ >= minZ && this.minZ <= maxZ && this.maxY >= minY && this.minY <= maxY;
+        return this.maxX >= other.minX
+            && this.minX <= other.maxX
+            && this.maxZ >= other.minZ
+            && this.minZ <= other.maxZ
+            && this.maxY >= other.minY
+            && this.minY <= other.maxY;
     }
 
     public boolean intersects(int minX, int minZ, int maxX, int maxZ) {
