--- a/net/minecraft/world/level/levelgen/structure/StructureSet.java
+++ b/net/minecraft/world/level/levelgen/structure/StructureSet.java
@@ -5,7 +_,7 @@
 import java.util.List;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
+import net.minecraft.resources.RegistryFileCodec;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.world.level.levelgen.structure.placement.StructurePlacement;
 
@@ -17,7 +_,7 @@
             )
             .apply(i, StructureSet::new)
     );
-    public static final Codec<Holder<StructureSet>> CODEC = RegistryCodecs.holder(Registries.STRUCTURE_SET, DIRECT_CODEC);
+    public static final Codec<Holder<StructureSet>> CODEC = RegistryFileCodec.create(Registries.STRUCTURE_SET, DIRECT_CODEC);
 
     public StructureSet(Holder<Structure> singleEntry, StructurePlacement placement) {
         this(List.of(new StructureSet.StructureSelectionEntry(singleEntry, 1)), placement);
