--- a/net/minecraft/world/level/levelgen/structure/StructureStart.java
+++ b/net/minecraft/world/level/levelgen/structure/StructureStart.java
@@ -17,6 +_,10 @@
 import net.minecraft.world.level.levelgen.structure.pieces.PiecesContainer;
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceSerializationContext;
 import net.minecraft.world.level.levelgen.structure.structures.OceanMonumentStructure;
+import org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry;
+import org.bukkit.craftbukkit.persistence.DirtyCraftPersistentDataContainer;
+import org.bukkit.craftbukkit.util.CraftStructureTransformer;
+import org.bukkit.craftbukkit.util.TransformerGeneratorAccess;
 import org.slf4j.Logger;
 
 public final class StructureStart {
@@ -30,6 +_,12 @@
     @Nullable
     private volatile BoundingBox cachedBoundingBox;
 
+    // CraftBukkit start
+    private static final CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new CraftPersistentDataTypeRegistry();
+    public DirtyCraftPersistentDataContainer persistentDataContainer = new DirtyCraftPersistentDataContainer(DATA_TYPE_REGISTRY);
+    public org.bukkit.event.world.AsyncStructureGenerateEvent.Cause generationEventCause = org.bukkit.event.world.AsyncStructureGenerateEvent.Cause.WORLD_GENERATION;
+    // CraftBukkit end
+
     public StructureStart(Structure p_226846_, ChunkPos p_226847_, int p_226848_, PiecesContainer p_226849_) {
         this.structure = p_226846_;
         this.chunkPos = p_226847_;
@@ -78,6 +_,7 @@
         return boundingbox;
     }
 
+    public TransformerGeneratorAccess placeInChunk$transformerAccess;
     public void placeInChunk(
         WorldGenLevel p_226851_, StructureManager p_226852_, ChunkGenerator p_226853_, RandomSource p_226854_, BoundingBox p_226855_, ChunkPos p_226856_
     ) {
@@ -87,11 +_,15 @@
             BlockPos blockpos = boundingbox.getCenter();
             BlockPos blockpos1 = new BlockPos(blockpos.getX(), boundingbox.minY(), blockpos.getZ());
 
+            placeInChunk$transformerAccess = new TransformerGeneratorAccess();
+            placeInChunk$transformerAccess.setHandle(p_226851_);
+            placeInChunk$transformerAccess.setStructureTransformer(new CraftStructureTransformer(generationEventCause, p_226851_, p_226852_, structure, p_226855_, p_226856_));
             for (StructurePiece structurepiece : list) {
                 if (structurepiece.getBoundingBox().intersects(p_226855_)) {
-                    structurepiece.postProcess(p_226851_, p_226852_, p_226853_, p_226854_, p_226855_, p_226856_, blockpos1);
+                    structurepiece.postProcess(placeInChunk$transformerAccess, p_226852_, p_226853_, p_226854_, p_226855_, p_226856_, blockpos1);
                 }
             }
+            placeInChunk$transformerAccess.getStructureTransformer().discard();
 
             this.structure.afterPlace(p_226851_, p_226852_, p_226853_, p_226854_, p_226855_, p_226856_, this.pieceContainer);
         }
@@ -99,7 +_,16 @@
 
     public CompoundTag createTag(StructurePieceSerializationContext p_192661_, ChunkPos p_192662_) {
         CompoundTag compoundtag = new CompoundTag();
+        // CraftBukkit start - store persistent data in nbt
+        if (!persistentDataContainer.isEmpty()) {
+            compoundtag.put("StructureBukkitValues", persistentDataContainer.toTagCompound());
+        }
+        // CraftBukkit end
+
         if (this.isValid()) {
+            if (p_192661_.registryAccess().registryOrThrow(Registries.STRUCTURE).getKey(this.getStructure()) == null) { // FORGE: This is just a more friendly error instead of the 'Null String' below
+                throw new RuntimeException("StructureStart \"" + this.getClass().getName() + "\": \"" + this.getStructure() + "\" unregistered, serializing impossible.");
+            }
             compoundtag.putString("id", p_192661_.registryAccess().registryOrThrow(Registries.STRUCTURE).getKey(this.structure).toString());
             compoundtag.putInt("ChunkX", p_192662_.x);
             compoundtag.putInt("ChunkZ", p_192662_.z);
