--- a/net/minecraft/world/level/levelgen/structure/StructurePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/StructurePiece.java
@@ -41,14 +_,14 @@
 public abstract class StructurePiece {
     private static final Logger LOGGER = LogUtils.getLogger();
     protected static final BlockState CAVE_AIR = Blocks.CAVE_AIR.defaultBlockState();
-    public BoundingBox boundingBox;
+    protected BoundingBox boundingBox;
     @Nullable
     private Direction orientation;
     private Mirror mirror;
     private Rotation rotation;
     protected int genDepth;
     private final StructurePieceType type;
-    public static final Set<Block> SHAPE_CHECK_BLOCKS = ImmutableSet.<Block>builder()
+    private static final Set<Block> SHAPE_CHECK_BLOCKS = ImmutableSet.<Block>builder()
         .add(Blocks.NETHER_BRICK_FENCE)
         .add(Blocks.TORCH)
         .add(Blocks.WALL_TORCH)
@@ -91,6 +_,9 @@
     }
 
     public final CompoundTag createTag(StructurePieceSerializationContext p_192645_) {
+        if (BuiltInRegistries.STRUCTURE_PIECE.getKey(this.getType()) == null) { // FORGE: Friendlier error then the Null String error below.
+            throw new RuntimeException("StructurePiece \"" + this.getClass().getName() + "\": \"" + this.getType() + "\" unregistered, serializing impossible.");
+        }
         CompoundTag compoundtag = new CompoundTag();
         compoundtag.putString("id", BuiltInRegistries.STRUCTURE_PIECE.getKey(this.getType()).toString());
         BoundingBox.CODEC
