--- a/net/minecraft/client/model/monster/illager/IllagerModel.java
+++ b/net/minecraft/client/model/monster/illager/IllagerModel.java
@@ -121,9 +_,8 @@
         if (pose == AbstractIllager.IllagerArmPose.ATTACKING) {
             if (state.getMainHandItemState().isEmpty()) {
                 AnimationUtils.animateZombieArms(this.leftArm, this.rightArm, true, state);
-                AnimationUtils.bobArms(this.rightArm, this.leftArm, state.ageInTicks);
             } else {
-                AnimationUtils.swingWeaponDown(this.rightArm, this.leftArm, state.mainArm, state.swingAnimation, state.ageInTicks);
+                AnimationUtils.swingWeaponDown(this.rightArm, this.leftArm, state.mainArm, state.attackAnim, state.ageInTicks);
             }
         } else if (pose == AbstractIllager.IllagerArmPose.SPELLCASTING) {
             this.rightArm.z = 0.0F;
