--- a/net/minecraft/world/entity/monster/breeze/Breeze.java
+++ b/net/minecraft/world/entity/monster/breeze/Breeze.java
@@ -206,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.BREEZE_DEATH;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_312252_) {
+    public SoundEvent getHurtSound(DamageSource p_312252_) {
         return SoundEvents.BREEZE_HURT;
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return this.onGround() ? SoundEvents.BREEZE_IDLE_GROUND : SoundEvents.BREEZE_IDLE_AIR;
     }
 
