--- a/net/minecraft/world/entity/monster/breeze/Breeze.java
+++ b/net/minecraft/world/entity/monster/breeze/Breeze.java
@@ -79,6 +_,28 @@
         this.xpReward = 10;
     }
 
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.breezeTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.breezeAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.breezeMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.breezeScale);
+    }
+    // Purpur end - Configurable entity base attributes
+
     @Override
     public Brain<Breeze> getBrain() {
         return (Brain<Breeze>) super.getBrain();
@@ -243,6 +_,10 @@
 
     @Override
     public boolean canAttack(LivingEntity target) {
+        // CraftBukkit start - SPIGOT-7957: Allow attack if target from brain was set
+        LivingEntity brainTarget = this.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).orElse(null);
+        if (brainTarget != null) return target.is(brainTarget);
+        // CraftBukkit end
         return (target.is(EntityTypes.PLAYER) || target.is(EntityTypes.IRON_GOLEM)) && super.canAttack(target);
     }
 
