--- a/net/minecraft/world/level/portal/PortalForcer.java
+++ b/net/minecraft/world/level/portal/PortalForcer.java
@@ -11,6 +_,7 @@
 import net.minecraft.world.entity.ai.village.poi.PoiManager;
 import net.minecraft.world.entity.ai.village.poi.PoiRecord;
 import net.minecraft.world.entity.ai.village.poi.PoiTypes;
+import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.NetherPortalBlock;
 import net.minecraft.world.level.block.state.BlockState;
@@ -39,17 +_,63 @@
     }
 
     public Optional<BlockPos> findClosestPortalPosition(BlockPos approximateExitPos, boolean toNether, WorldBorder worldBorder) {
+        // CraftBukkit start
+        return findClosestPortalPosition(approximateExitPos, worldBorder, toNether ? 16 : 128); // Search Radius
+    }
+
+    public Optional<BlockPos> findClosestPortalPosition(BlockPos approximateExitPos, WorldBorder worldBorder, int radius) {
         PoiManager poiManager = this.level.getPoiManager();
-        int radius = toNether ? 16 : 128;
-        poiManager.ensureLoadedAndValid(this.level, approximateExitPos, radius);
-        return poiManager.getInSquare(type -> type.is(PoiTypes.NETHER_PORTAL), approximateExitPos, radius, PoiManager.Occupancy.ANY)
-            .map(PoiRecord::getPos)
-            .filter(worldBorder::isWithinBounds)
-            .filter(pos -> this.level.getBlockState(pos).hasProperty(BlockStateProperties.HORIZONTAL_AXIS))
-            .min(Comparator.<BlockPos>comparingDouble(p -> p.distSqr(approximateExitPos)).thenComparingInt(Vec3i::getY));
+        // int radius = toNether ? 16 : 128;
+        // CraftBukkit end
+        // Paper start - optimise poi lookup
+        //poiManager.ensureLoadedAndValid(this.level, approximateExitPos, radius); // Paper - optimise poi lookup - we can let it load as it reads
+        java.util.List<PoiRecord> records = new java.util.ArrayList<>();
+        ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findClosestPoiDataRecords(
+                poiManager, type -> type.is(PoiTypes.NETHER_PORTAL),
+                (final net.minecraft.core.Holder<net.minecraft.world.entity.ai.village.poi.PoiType> type, final BlockPos pos) -> {
+                    if (!worldBorder.isWithinBounds(pos)) {
+                        return false;
+                    }
+                    // Paper start - Configurable nether ceiling damage
+                    if (this.level.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER && this.level.paperConfig().environment.netherCeilingVoidDamageHeight.test(v -> pos.getY() >= v)) {
+                        return false;
+                    }
+                    // Paper end - Configurable nether ceiling damage
+
+                    net.minecraft.world.level.chunk.ChunkAccess lowest = this.level.getChunk(pos.getX() >> 4, pos.getZ() >> 4, net.minecraft.world.level.chunk.status.ChunkStatus.EMPTY);
+                    net.minecraft.world.level.levelgen.BelowZeroRetrogen belowZeroRetrogen;
+                    if (!lowest.getPersistedStatus().isOrAfter(net.minecraft.world.level.chunk.status.ChunkStatus.FULL)
+                            // check below zero retrogen so that pre 1.17 worlds still load portals (JMP)
+                            && ((belowZeroRetrogen = lowest.getBelowZeroRetrogen()) == null || !belowZeroRetrogen.targetStatus().isOrAfter(net.minecraft.world.level.chunk.status.ChunkStatus.SPAWN))) {
+                        // why would we generate the chunk?
+                        return false;
+                    }
+                    return lowest.getBlockState(pos).hasProperty(BlockStateProperties.HORIZONTAL_AXIS);
+                },
+                approximateExitPos, radius, Double.MAX_VALUE, PoiManager.Occupancy.ANY, true, records
+        );
+
+        // this gets us most of the way there, but Vanilla biases lower y values.
+        PoiRecord lowestYRecord = null;
+        for (PoiRecord record : records) {
+            if (lowestYRecord == null) {
+                lowestYRecord = record;
+            } else if (lowestYRecord.getPos().getY() > record.getPos().getY()) {
+                lowestYRecord = record;
+            }
+        }
+        // now we're done
+        return Optional.ofNullable(lowestYRecord == null ? null : lowestYRecord.getPos());
+        // Paper end - optimise poi lookup
     }
 
     public Optional<BlockUtil.FoundRectangle> createPortal(BlockPos origin, Direction.Axis portalAxis) {
+        // CraftBukkit start
+        return this.createPortal(origin, portalAxis, null, 16);
+    }
+
+    public Optional<BlockUtil.FoundRectangle> createPortal(BlockPos origin, Direction.Axis portalAxis, net.minecraft.world.entity.Entity entity, int createRadius) {
+        // CraftBukkit end
         Direction direction = Direction.get(Direction.AxisDirection.POSITIVE, portalAxis);
         double closestFullDistanceSqr = -1.0;
         BlockPos closestFullPosition = null;
@@ -57,10 +_,15 @@
         BlockPos closestPartialPosition = null;
         WorldBorder worldBorder = this.level.getWorldBorder();
         int maxPlaceableY = Math.min(this.level.getMaxY(), this.level.getMinY() + this.level.getLogicalHeight() - 1);
+        // Paper start - Configurable nether ceiling damage; make sure the max height doesn't exceed the void damage height
+        if (this.level.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER && this.level.paperConfig().environment.netherCeilingVoidDamageHeight.enabled()) {
+            maxPlaceableY = Math.min(maxPlaceableY, this.level.paperConfig().environment.netherCeilingVoidDamageHeight.intValue() - 1);
+        }
+        // Paper end - Configurable nether ceiling damage
         int edgeDistance = 1;
         BlockPos.MutableBlockPos mutable = origin.mutable();
 
-        for (BlockPos.MutableBlockPos columnPos : BlockPos.spiralAround(origin, 16, Direction.EAST, Direction.SOUTH)) {
+        for (BlockPos.MutableBlockPos columnPos : BlockPos.spiralAround(origin, createRadius, Direction.EAST, Direction.SOUTH)) { // CraftBukkit
             int height = Math.min(maxPlaceableY, this.level.getHeight(Heightmap.Types.MOTION_BLOCKING, columnPos.getX(), columnPos.getZ()));
             if (worldBorder.isWithinBounds(columnPos) && worldBorder.isWithinBounds(columnPos.move(direction, 1))) {
                 columnPos.move(direction.getOpposite(), 1);
@@ -104,6 +_,7 @@
             closestFullDistanceSqr = closestPartialDistanceSqr;
         }
 
+        org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(this.level); // CraftBukkit - Use BlockStateListPopulator
         if (closestFullDistanceSqr == -1.0) {
             int minStartY = Math.max(this.level.getMinY() - -1, 70);
             int maxStartY = maxPlaceableY - 9;
@@ -128,7 +_,7 @@
                             height,
                             width * direction.getStepZ() + box * clockWise.getStepZ()
                         );
-                        this.level.setBlockAndUpdate(mutable, blockState);
+                        blockList.setBlock(mutable, blockState, Block.UPDATE_ALL); // CraftBukkit
                     }
                 }
             }
@@ -138,7 +_,7 @@
             for (int height = -1; height < 4; height++) {
                 if (width == -1 || width == 2 || height == -1 || height == 3) {
                     mutable.setWithOffset(closestFullPosition, width * direction.getStepX(), height, width * direction.getStepZ());
-                    this.level.setBlockAndUpdate(mutable, Blocks.OBSIDIAN.defaultBlockState());
+                    blockList.setBlock(mutable, Blocks.OBSIDIAN.defaultBlockState(), Block.UPDATE_ALL); // CraftBukkit
                 }
             }
         }
@@ -146,12 +_,22 @@
         BlockState portalBlockState = Blocks.NETHER_PORTAL.defaultBlockState().setValue(NetherPortalBlock.AXIS, portalAxis);
 
         for (int width = 0; width < 2; width++) {
-            for (int height = 0; height < 3; height++) {
-                mutable.setWithOffset(closestFullPosition, width * direction.getStepX(), height, width * direction.getStepZ());
-                this.level.setBlock(mutable, portalBlockState, 18);
+            for (int heightx = 0; heightx < 3; heightx++) {
+                mutable.setWithOffset(closestFullPosition, width * direction.getStepX(), heightx, width * direction.getStepZ());
+                blockList.setBlock(mutable, portalBlockState, Block.UPDATE_CLIENTS | Block.UPDATE_KNOWN_SHAPE); // CraftBukkit
             }
         }
 
+        // CraftBukkit start
+        org.bukkit.World bworld = this.level.getWorld();
+        org.bukkit.event.world.PortalCreateEvent event = new org.bukkit.event.world.PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blockList.getSnapshotBlocks(), bworld, entity == null ? null : entity.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.NETHER_PAIR);
+
+        this.level.getCraftServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            return Optional.empty();
+        }
+        blockList.placeBlocks();
+        // CraftBukkit end
         return Optional.of(new BlockUtil.FoundRectangle(closestFullPosition.immutable(), 2, 3));
     }
 
@@ -168,6 +_,13 @@
                 mutable.setWithOffset(
                     origin, direction.getStepX() * width + clockWise.getStepX() * offset, height, direction.getStepZ() * width + clockWise.getStepZ() * offset
                 );
+                // Paper start - Protect Bedrock and End Portal/Frames from being destroyed
+                if (!io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPermanentBlockBreakExploits) {
+                    if (!this.level.getBlockState(mutable).isDestroyable()) {
+                        return false;
+                    }
+                }
+                // Paper end - Protect Bedrock and End Portal/Frames from being destroyed
                 if (height < 0 && !this.level.getBlockState(mutable).isSolid()) {
                     return false;
                 }
