--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -94,11 +_,16 @@
     public ChunkGenerator(BiomeSource biomeSource, Function<Holder<Biome>, BiomeGenerationSettings> generationSettingsGetter) {
         this.biomeSource = biomeSource;
         this.generationSettingsGetter = generationSettingsGetter;
-        this.featuresPerStep = Suppliers.memoize(
+        this.featuresPerStep = net.neoforged.neoforge.common.util.Lazy.of(
             () -> FeatureSorter.buildFeaturesPerStep(List.copyOf(biomeSource.possibleBiomes()), b -> generationSettingsGetter.apply(b).features(), true)
         );
     }
 
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public void refreshFeaturesPerStep() {
+        ((net.neoforged.neoforge.common.util.Lazy<?>) this.featuresPerStep).invalidate();
+    }
+
     public void validate() {
         this.featuresPerStep.get();
     }
@@ -117,7 +_,7 @@
         return CompletableFuture.supplyAsync(() -> {
             protoChunk.fillBiomesFromNoise(this.biomeSource, randomState.sampler());
             return protoChunk;
-        }, Util.backgroundExecutor().forName("init_biomes"));
+        }, Runnable::run); // Paper - rewrite chunk system
     }
 
     public abstract void applyCarvers(
@@ -294,7 +_,7 @@
                     return Pair.of(config.getLocatePos(chunkTarget), structure);
                 }
 
-                ChunkAccess chunk = level.getChunk(chunkTarget.x(), chunkTarget.z(), ChunkStatus.STRUCTURE_STARTS);
+                ChunkAccess chunk = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevelReader)level).moonrise$syncLoadNonFull(chunkTarget.x(), chunkTarget.z(), ChunkStatus.STRUCTURE_STARTS); // Paper - rewrite chunk system
                 StructureStart start = structureManager.getStartForStructure(SectionPos.bottomOf(chunk), structure.value(), chunk);
                 if (start != null && start.isValid() && (!createReference || tryAddReference(structureManager, start))) {
                     return Pair.of(config.getLocatePos(start.getChunkPos()), structure);
@@ -416,6 +_,33 @@
             }
         }
     }
+
+    // CraftBukkit start
+    public void applyBiomeDecorationCB(WorldGenLevel level, ChunkAccess chunk, StructureManager structureManager) {
+        applyBiomeDecorationCB(level, chunk, structureManager, true);
+    }
+
+    public void applyBiomeDecorationCB(WorldGenLevel worldgenlevel, ChunkAccess chunkaccess, StructureManager structuremanager, boolean vanilla) {
+        if (vanilla) {
+            applyBiomeDecoration(worldgenlevel, chunkaccess, structuremanager);
+        }
+
+        org.bukkit.World world = worldgenlevel.getMinecraftWorld().getWorld();
+        // only call when a populator is present (prevents unnecessary entity conversion)
+        if (!world.getPopulators().isEmpty()) {
+            org.bukkit.craftbukkit.generator.CraftLimitedRegion limitedRegion = new org.bukkit.craftbukkit.generator.CraftLimitedRegion(worldgenlevel, chunkaccess.getPos());
+            int x = chunkaccess.getPos().x();
+            int z = chunkaccess.getPos().z();
+            for (org.bukkit.generator.BlockPopulator populator : world.getPopulators()) {
+                WorldgenRandom worldgenrandom = new WorldgenRandom(new net.minecraft.world.level.levelgen.LegacyRandomSource(worldgenlevel.getSeed()));
+                worldgenrandom.setDecorationSeed(worldgenlevel.getSeed(), x, z);
+                populator.populate(world, new org.bukkit.craftbukkit.util.RandomSourceWrapper.RandomWrapper(worldgenrandom), x, z, limitedRegion);
+            }
+            limitedRegion.saveEntities();
+            limitedRegion.breakLink();
+        }
+    }
+    // CraftBukkit end
 
     private static BoundingBox getWritableArea(ChunkAccess chunk) {
         ChunkPos chunkPos = chunk.getPos();
