--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.level.chunk;
 
+import java.util.Objects;
 import java.util.function.Predicate;
 import net.minecraft.core.Holder;
 import net.minecraft.network.FriendlyByteBuf;
@@ -16,7 +_,7 @@
     private short tickingBlockCount;
     private short tickingFluidCount;
     public final PalettedContainer<BlockState> states;
-    private PalettedContainerRO<Holder<Biome>> biomes;
+    private PalettedContainer<Holder<Biome>> biomes;
 
     private LevelChunkSection(LevelChunkSection source) {
         this.nonEmptyBlockCount = source.nonEmptyBlockCount;
@@ -27,14 +_,20 @@
         this.biomes = source.biomes.copy();
     }
 
-    public LevelChunkSection(PalettedContainer<BlockState> states, PalettedContainerRO<Holder<Biome>> biomes) {
+    public LevelChunkSection(PalettedContainer<BlockState> states, PalettedContainer<Holder<Biome>> biomes) {
         this.states = states;
         this.biomes = biomes;
         this.recalcBlockCounts();
     }
 
+    @Deprecated @io.papermc.paper.annotation.DoNotUse // Paper - Anti-Xray
     public LevelChunkSection(PalettedContainerFactory containerFactory) {
-        this.states = containerFactory.createForBlockStates();
+        // Paper start - Anti-Xray - Add parameters
+        this(containerFactory, null, null, 0);
+    }
+    public LevelChunkSection(final PalettedContainerFactory containerFactory, final net.minecraft.world.level.Level level, final net.minecraft.world.level.ChunkPos chunkPos, final int chunkSectionY) {
+        this.states = containerFactory.createForBlockStates(level, chunkPos, chunkSectionY);
+        // Paper end - Anti-Xray - Add parameters
         this.biomes = containerFactory.createForBiomes();
     }
 
@@ -66,7 +_,7 @@
             previous = this.states.getAndSetUnchecked(sectionX, sectionY, sectionZ, state);
         }
 
-        if (!previous.isAir()) {
+        if (!previous.isEmpty()) { // Neo: Fix MC-232360 for modded blocks (Makes modded isAir blocks not be replaced with Blocks.AIR in all-air chunk sections)
             this.nonEmptyBlockCount--;
             if (previous.isRandomlyTicking()) {
                 this.tickingBlockCount--;
@@ -81,7 +_,7 @@
             }
         }
 
-        if (!state.isAir()) {
+        if (!state.isEmpty()) { // Neo: Fix MC-232360 for modded blocks (Makes modded isAir blocks not be replaced with Blocks.AIR in all-air chunk sections)
             this.nonEmptyBlockCount++;
             if (state.isRandomlyTicking()) {
                 this.tickingBlockCount++;
@@ -127,7 +_,7 @@
             public int tickingFluidCount;
 
             public void accept(BlockState state, int count) {
-                if (!state.isAir()) {
+                if (!state.isEmpty()) { // Neo: Fix MC-232360 for modded blocks (Makes modded isAir blocks not be replaced with Blocks.AIR in all-air chunk sections)
                     this.nonEmptyBlockCount += count;
                     if (state.isRandomlyTicking()) {
                         this.tickingBlockCount += count;
@@ -193,6 +_,12 @@
     public Holder<Biome> getNoiseBiome(int quartX, int quartY, int quartZ) {
         return this.biomes.get(quartX, quartY, quartZ);
     }
+
+    // CraftBukkit start
+    public void setNoiseBiome(int quartX, int quartY, int quartZ, Holder<Biome> biome) {
+        this.biomes.set(quartX, quartY, quartZ, biome);
+    }
+    // CraftBukkit end
 
     public void fillBiomesFromNoise(BiomeResolver biomeResolver, Climate.Sampler sampler, int quartMinX, int quartMinY, int quartMinZ) {
         PalettedContainer<Holder<Biome>> newBiomes = this.biomes.recreate();
