--- a/net/minecraft/world/level/chunk/PalettedContainerFactory.java
+++ b/net/minecraft/world/level/chunk/PalettedContainerFactory.java
@@ -17,7 +_,8 @@
     Codec<PalettedContainer<BlockState>> blockStatesContainerCodec,
     Strategy<Holder<Biome>> biomeStrategy,
     Holder<Biome> defaultBiome,
-    Codec<PalettedContainerRO<Holder<Biome>>> biomeContainerCodec
+    Codec<PalettedContainerRO<Holder<Biome>>> biomeContainerCodec,
+    Codec<PalettedContainer<Holder<Biome>>> biomeContainerCodecRW
 ) {
     public static PalettedContainerFactory create(RegistryAccess registries) {
         Strategy<BlockState> blockStateStrategy = Strategy.createForBlockStates(Block.BLOCK_STATE_REGISTRY);
@@ -31,7 +_,8 @@
             PalettedContainer.codecRW(BlockState.CODEC, blockStateStrategy, defaultBlockState),
             biomeStrategy,
             defaultBiome,
-            PalettedContainer.codecRO(biomes.holderByNameCodec(), biomeStrategy, defaultBiome)
+            PalettedContainer.codecRO(biomes.holderByNameCodec(), biomeStrategy, defaultBiome),
+            PalettedContainer.codecRW(biomes.holderByNameCodec(), biomeStrategy, defaultBiome)
         );
     }
 
