--- a/net/minecraft/world/level/chunk/PalettedContainerRO.java
+++ b/net/minecraft/world/level/chunk/PalettedContainerRO.java
@@ -14,7 +_,11 @@
 
     void getAll(Consumer<T> consumer);
 
+    // Paper start - Anti-Xray - Add chunk packet info
+    @Deprecated @io.papermc.paper.annotation.DoNotUse
     void write(FriendlyByteBuf buffer);
+    void write(FriendlyByteBuf buffer, io.papermc.paper.antixray.@org.jspecify.annotations.Nullable ChunkPacketInfo<T> chunkPacketInfo, int chunkSectionIndex);
+    // Paper end
 
     int getSerializedSize();
 
