--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -45,6 +_,7 @@
     private volatile @Nullable LevelLightEngine lightEngine;
     private volatile ChunkStatus status = ChunkStatus.EMPTY;
     private final List<CompoundTag> entities = Lists.newArrayList();
+    private @Nullable CarvingMask carvingMask;
     private @Nullable BelowZeroRetrogen belowZeroRetrogen;
     private final ProtoChunkTicks<Block> blockTicks;
     private final ProtoChunkTicks<Fluid> fluidTicks;
@@ -89,15 +_,32 @@
         return new ChunkAccess.PackedTicks(this.blockTicks.pack(currentTick), this.fluidTicks.pack(currentTick));
     }
 
+    // Paper start - If loaded util
+    @Override
+    public final FluidState getFluidIfLoaded(BlockPos pos) {
+        return this.getFluidState(pos);
+    }
+
+    @Override
+    public final BlockState getBlockStateIfLoaded(BlockPos pos) {
+        return this.getBlockState(pos);
+    }
+    // Paper end
+
     @Override
     public BlockState getBlockState(BlockPos pos) {
-        int y = pos.getY();
+        // Paper start
+        return this.getBlockState(pos.getX(), pos.getY(), pos.getZ());
+    }
+
+    public BlockState getBlockState(final int x, final int y, final int z) {
+        // Paper end
         if (this.isOutsideBuildHeight(y)) {
             return Blocks.VOID_AIR.defaultBlockState();
         }
 
         LevelChunkSection section = this.getSection(this.getSectionIndex(y));
-        return section.hasOnlyAir() ? Blocks.AIR.defaultBlockState() : section.getBlockState(pos.getX() & 15, y & 15, pos.getZ() & 15);
+        return section.hasOnlyAir() ? Blocks.AIR.defaultBlockState() : section.getBlockState(x & 15, y & 15, z & 15); // Paper
     }
 
     @Override
@@ -137,8 +_,8 @@
                 this.lightEngine.updateSectionStatus(pos, isEmpty);
             }
 
-            if (LightEngine.hasDifferentLightProperties(oldState, state)) {
-                this.skyLightSources.update(this, localX, y, localZ);
+            if (LightEngine.hasDifferentLightProperties(this, pos, oldState, state)) {
+                // Paper - rewrite chunk system
                 this.lightEngine.checkBlock(pos);
             }
         }
@@ -154,8 +_,6 @@
                 }
 
                 toPrime.add(type);
-            } else {
-                heightmap.update(localX, y, localZ, state);
             }
         }
 
@@ -163,6 +_,10 @@
             Heightmap.primeHeightmaps(this, toPrime);
         }
 
+        for (Heightmap.Types type : heightmapsAfter) {
+            this.heightmaps.get(type).update(localX, y, localZ, state);
+        }
+
         return oldState;
     }
 
@@ -280,6 +_,22 @@
     public void removeBlockEntity(BlockPos pos) {
         this.blockEntities.remove(pos);
         this.pendingBlockEntities.remove(pos);
+    }
+
+    public @Nullable CarvingMask getCarvingMask() {
+        return this.carvingMask;
+    }
+
+    public CarvingMask getOrCreateCarvingMask() {
+        if (this.carvingMask == null) {
+            this.carvingMask = new CarvingMask(this.getHeight(), this.getMinY());
+        }
+
+        return this.carvingMask;
+    }
+
+    public void setCarvingMask(CarvingMask data) {
+        this.carvingMask = data;
     }
 
     public void setLightEngine(LevelLightEngine lightEngine) {
