--- a/net/minecraft/world/level/chunk/Strategy.java
+++ b/net/minecraft/world/level/chunk/Strategy.java
@@ -4,7 +_,7 @@
 import net.minecraft.util.Mth;
 
 public abstract class Strategy<T> {
-    private static final Palette.Factory SINGLE_VALUE_PALETTE_FACTORY = SingleValuePalette::create;
+    public static final Palette.Factory SINGLE_VALUE_PALETTE_FACTORY = SingleValuePalette::create; // Paper - Anti-Xray
     private static final Palette.Factory LINEAR_PALETTE_FACTORY = LinearPalette::create;
     private static final Palette.Factory HASHMAP_PALETTE_FACTORY = HashMapPalette::create;
     private static final Configuration ZERO_BITS = new Configuration.Simple(SINGLE_VALUE_PALETTE_FACTORY, 0);
