--- a/net/minecraft/client/sounds/SoundEngine.java
+++ b/net/minecraft/client/sounds/SoundEngine.java
@@ -56,7 +_,7 @@
     public static final String MISSING_SOUND = "FOR THE DEBUG!";
     public static final String OPEN_AL_SOFT_PREFIX = "OpenAL Soft on ";
     public static final int OPEN_AL_SOFT_PREFIX_LENGTH = "OpenAL Soft on ".length();
-    public final SoundManager soundManager;
+    private final SoundManager soundManager;
     private final Options options;
     private boolean loaded;
     private final Library library = new Library();
