--- a/net/minecraft/client/sounds/SoundEngine.java
+++ b/net/minecraft/client/sounds/SoundEngine.java
@@ -54,7 +_,6 @@
     private static final int MIN_SOURCE_LIFETIME = 20;
     private static final Set<Identifier> ONLY_WARN_ONCE = Sets.newHashSet();
     public static final String MISSING_SOUND = "FOR THE DEBUG!";
-    public static final int LOOPING_SOUND_SUBTITLE_INTERVAL_TICKS = 10;
     public static final String OPEN_AL_SOFT_PREFIX = "OpenAL Soft on ";
     public static final int OPEN_AL_SOFT_PREFIX_LENGTH = "OpenAL Soft on ".length();
     private final SoundManager soundManager;
@@ -254,16 +_,6 @@
                         channel.setPitch(pitch);
                         channel.setSelfPosition(position);
                     });
-                    if (instance.isLooping() && this.tickCount % 10 == 0 && volume > 0.0F && !this.listeners.isEmpty()) {
-                        Sound sound = instance.getSound();
-                        if (sound != null) {
-                            WeighedSoundEvents soundEvent = instance.getSoundEvent();
-                            if (soundEvent != null) {
-                                float attenuationDistance = sound.getAttenuationDistance(instance.getVolume());
-                                this.notifyListeners(instance, soundEvent, attenuationDistance);
-                            }
-                        }
-                    }
                 }
             }
         }
@@ -287,7 +_,7 @@
 
                     try {
                         this.instanceBySource.remove(instance.getSource(), instance);
-                    } catch (RuntimeException var10) {
+                    } catch (RuntimeException var7) {
                     }
 
                     if (instance instanceof TickableSoundInstance) {
@@ -313,18 +_,6 @@
         }
     }
 
-    private void notifyListeners(SoundInstance instance, WeighedSoundEvents soundEvent, float attenuationDistance) {
-        float range = getRange(instance.isRelative(), instance.getAttenuation(), attenuationDistance);
-
-        for (SoundEventListener listener : this.listeners) {
-            listener.onPlaySound(instance, soundEvent, range);
-        }
-    }
-
-    private static float getRange(boolean isRelative, SoundInstance.Attenuation instance, float attenuationDistance) {
-        return !isRelative && instance != SoundInstance.Attenuation.NONE ? attenuationDistance : Float.POSITIVE_INFINITY;
-    }
-
     private void tickMusicWhenPaused() {
         Iterator<Entry<SoundInstance, ChannelAccess.ChannelHandle>> iterator = this.instanceToChannel.entrySet().iterator();
 
@@ -372,7 +_,7 @@
             return SoundEngine.PlayResult.NOT_STARTED;
         }
 
-        WeighedSoundEvents soundEvent = instance.getOrResolve(this.soundManager);
+        WeighedSoundEvents soundEvent = instance.resolve(this.soundManager);
         Identifier eventLocation = instance.getIdentifier();
         if (soundEvent == null) {
             if (ONLY_WARN_ONCE.add(eventLocation)) {
@@ -399,13 +_,20 @@
             return SoundEngine.PlayResult.NOT_STARTED;
         } else {
             float instanceVolume = instance.getVolume();
-            float attenuationDistance = sound.getAttenuationDistance(instanceVolume);
-            if (!this.listeners.isEmpty()) {
-                this.notifyListeners(instance, soundEvent, attenuationDistance);
-            }
-
+            float attenuationDistance = Math.max(instanceVolume, 1.0F) * sound.getAttenuationDistance();
             SoundSource soundSource = instance.getSource();
             float volume = this.calculateVolume(instanceVolume, soundSource);
+            float pitch = this.calculatePitch(instance);
+            SoundInstance.Attenuation attenuation = instance.getAttenuation();
+            boolean isRelative = instance.isRelative();
+            if (!this.listeners.isEmpty()) {
+                float range = !isRelative && attenuation != SoundInstance.Attenuation.NONE ? attenuationDistance : Float.POSITIVE_INFINITY;
+
+                for (SoundEventListener listener : this.listeners) {
+                    listener.onPlaySound(instance, soundEvent, range);
+                }
+            }
+
             boolean startedSilently = false;
             if (volume == 0.0F) {
                 if (!instance.canStartSilent() && soundSource != SoundSource.MUSIC) {
@@ -433,11 +_,10 @@
                 this.soundDeleteTime.put(instance, this.tickCount + 20);
                 this.instanceToChannel.put(instance, handle);
                 this.instanceBySource.put(soundSource, instance);
-                float pitch = this.calculatePitch(instance);
                 handle.execute(channel -> {
                     channel.setPitch(pitch);
                     channel.setVolume(volume);
-                    if (instance.getAttenuation() == SoundInstance.Attenuation.LINEAR) {
+                    if (attenuation == SoundInstance.Attenuation.LINEAR) {
                         channel.linearAttenuation(attenuationDistance);
                     } else {
                         channel.disableAttenuation();
@@ -445,7 +_,7 @@
 
                     channel.setLooping(isLooping && !isStreaming);
                     channel.setSelfPosition(position);
-                    channel.setRelative(instance.isRelative());
+                    channel.setRelative(isRelative);
                 });
                 if (!isStreaming) {
                     this.soundBuffers.getCompleteBuffer(sound.getPath()).thenAccept(soundBuffer -> handle.execute(channel -> {
