--- /dev/null
+++ b/net/minecraft/world/level/storage/loot/providers/number/ConditionalValue.java
@@ -1,0 +_,49 @@
+package net.minecraft.world.level.storage.loot.providers.number;
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.MapCodec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import net.minecraft.core.Holder;
+import net.minecraft.world.level.storage.loot.LootContext;
+import net.minecraft.world.level.storage.loot.Validatable;
+import net.minecraft.world.level.storage.loot.ValidationContext;
+import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
+
+public record ConditionalValue(Holder<LootItemCondition> condition, Holder<NumberProvider> onTrue, Holder<NumberProvider> onFalse) implements NumberProvider {
+    public static final MapCodec<ConditionalValue> MAP_CODEC = RecordCodecBuilder.mapCodec(
+        i -> i.group(
+                LootItemCondition.CODEC.fieldOf("condition").forGetter(c -> c.condition),
+                NumberProviders.CODEC.fieldOf("on_true").forGetter(c -> c.onTrue),
+                NumberProviders.CODEC.optionalFieldOf("on_false", ConstantValue.exactly(0.0F)).forGetter(c -> c.onFalse)
+            )
+            .apply(i, ConditionalValue::new)
+    );
+    public static final Codec<ConditionalValue> CODEC = MAP_CODEC.codec();
+
+    @Override
+    public int getInt(LootContext context) {
+        return this.selectValue(context).getInt(context);
+    }
+
+    @Override
+    public float getFloat(LootContext context) {
+        return this.selectValue(context).getFloat(context);
+    }
+
+    private NumberProvider selectValue(LootContext context) {
+        return !this.condition.value().test(context) ? this.onFalse.value() : this.onTrue.value();
+    }
+
+    @Override
+    public MapCodec<ConditionalValue> codec() {
+        return MAP_CODEC;
+    }
+
+    @Override
+    public void validate(ValidationContext context) {
+        NumberProvider.super.validate(context);
+        Validatable.validateHolder(context, "condition", this.condition);
+        Validatable.validateHolder(context, "on_true", this.onTrue);
+        Validatable.validateHolder(context, "on_false", this.onFalse);
+    }
+}
