--- /dev/null
+++ b/net/minecraft/world/level/storage/loot/providers/number/NumberProvider.java
@@ -1,0 +_,15 @@
+package net.minecraft.world.level.storage.loot.providers.number;
+
+import com.mojang.serialization.MapCodec;
+import net.minecraft.world.level.storage.loot.LootContext;
+import net.minecraft.world.level.storage.loot.LootContextUser;
+
+public interface NumberProvider extends LootContextUser {
+    float getFloat(LootContext context);
+
+    default int getInt(LootContext context) {
+        return Math.round(this.getFloat(context));
+    }
+
+    MapCodec<? extends NumberProvider> codec();
+}
