--- /dev/null
+++ b/net/minecraft/world/level/storage/loot/providers/number/NumberProviderTypes.java
@@ -1,0 +_,20 @@
+package net.minecraft.world.level.storage.loot.providers.number;
+
+import com.mojang.serialization.MapCodec;
+import net.minecraft.core.Registry;
+
+public class NumberProviderTypes {
+    public static MapCodec<? extends NumberProvider> bootstrap(Registry<MapCodec<? extends NumberProvider>> registry) {
+        Registry.register(registry, "constant", ConstantValue.MAP_CODEC);
+        Registry.register(registry, "uniform", UniformGenerator.MAP_CODEC);
+        Registry.register(registry, "binomial", BinomialDistributionGenerator.MAP_CODEC);
+        Registry.register(registry, "score", ScoreboardValue.MAP_CODEC);
+        Registry.register(registry, "storage", StorageValue.MAP_CODEC);
+        Registry.register(registry, "sum", Sum.MAP_CODEC);
+        Registry.register(registry, "enchantment_level", EnchantmentLevelProvider.MAP_CODEC);
+        Registry.register(registry, "weighted_list", WeightedListValue.MAP_CODEC);
+        Registry.register(registry, "conditional", ConditionalValue.MAP_CODEC);
+        Registry.register(registry, "number_dispatcher", NumberDispatcher.MAP_CODEC);
+        return Registry.register(registry, "environment_attribute", EnvironmentAttributeValue.MAP_CODEC);
+    }
+}
