--- a/net/minecraft/world/level/storage/loot/providers/number/StorageValue.java
+++ b/net/minecraft/world/level/storage/loot/providers/number/StorageValue.java
@@ -3,7 +_,6 @@
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.List;
 import java.util.Optional;
 import net.minecraft.commands.arguments.NbtPathArgument;
