--- a/net/minecraft/locale/Language.java
+++ b/net/minecraft/locale/Language.java
@@ -7,12 +_,14 @@
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
+import com.mohistmc.youer.YouerConfig;
 import com.mojang.logging.LogUtils;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Map.Entry;
@@ -21,7 +_,6 @@
 import net.minecraft.network.chat.FormattedText;
 import net.minecraft.network.chat.Style;
 import net.minecraft.util.FormattedCharSequence;
-import net.minecraft.util.FormattedCharSink;
 import net.minecraft.util.GsonHelper;
 import net.minecraft.util.StringDecomposer;
 import org.slf4j.Logger;
@@ -36,8 +_,10 @@
     private static Language loadDefault() {
         Builder<String, String> builder = ImmutableMap.builder();
         BiConsumer<String, String> biconsumer = builder::put;
-        parseTranslations(biconsumer, "/assets/minecraft/lang/en_us.json");
-        final Map<String, String> map = builder.build();
+        Map<String, net.minecraft.network.chat.Component> componentMap = new java.util.HashMap<>();
+        parseTranslations(biconsumer, componentMap::put, "/assets/minecraft/lang/en_us.json");
+        final Map<String, String> map = new java.util.HashMap<>(builder.build());
+        net.neoforged.neoforge.server.LanguageHook.captureLanguageMap(map, componentMap);
         return new Language() {
             @Override
             public String getOrDefault(String p_128127_, String p_265421_) {
@@ -64,21 +_,55 @@
                         )
                         .isPresent();
             }
+
+            @Override
+            public Map<String, String> getLanguageData() {
+                return map;
+            }
+
+            @Override
+            public @org.jetbrains.annotations.Nullable net.minecraft.network.chat.Component getComponent(String key) {
+                return componentMap.get(key);
+            }
         };
     }
 
+    @Deprecated
     private static void parseTranslations(BiConsumer<String, String> p_282031_, String p_283638_) {
-        try (InputStream inputstream = Language.class.getResourceAsStream(p_283638_)) {
-            loadFromJson(inputstream, p_282031_);
-        } catch (JsonParseException | IOException ioexception) {
-            LOGGER.error("Couldn't read strings from {}", p_283638_, ioexception);
+        parseTranslations(p_282031_, (key, value) -> {}, p_283638_);
+    }
+
+    private static void parseTranslations(BiConsumer<String, String> p_282031_, BiConsumer<String, net.minecraft.network.chat.Component> componentConsumer, String p_283638_) {
+        try (InputStream inputstream = Language.class.getResourceAsStream("/assets/minecraft/lang/%s.json".formatted(YouerConfig.youer_lang().toLowerCase(Locale.ROOT)))) {
+            loadFromJson(inputstream, p_282031_, componentConsumer);
+        } catch (Exception ioexception) {
+            try (InputStream inputstream = Language.class.getResourceAsStream(p_283638_)) {
+                loadFromJson(inputstream, p_282031_, componentConsumer);
+            } catch (JsonParseException | IOException ioexception1) {
+                LOGGER.error("Couldn't read strings from {}", p_283638_, ioexception1);
+            }
         }
     }
 
     public static void loadFromJson(InputStream p_128109_, BiConsumer<String, String> p_128110_) {
+        loadFromJson(p_128109_, p_128110_, (key, value) -> {});
+    }
+
+    public static void loadFromJson(InputStream p_128109_, BiConsumer<String, String> p_128110_, BiConsumer<String, net.minecraft.network.chat.Component> componentConsumer) {
         JsonObject jsonobject = GSON.fromJson(new InputStreamReader(p_128109_, StandardCharsets.UTF_8), JsonObject.class);
 
         for (Entry<String, JsonElement> entry : jsonobject.entrySet()) {
+            if (entry.getValue().isJsonArray()) {
+                var component = net.minecraft.network.chat.ComponentSerialization.CODEC
+                    .parse(com.mojang.serialization.JsonOps.INSTANCE, entry.getValue())
+                    .getOrThrow(msg -> new com.google.gson.JsonParseException("Error parsing translation for " + entry.getKey() + ": " + msg));
+
+                p_128110_.accept(entry.getKey(), component.getString());
+                componentConsumer.accept(entry.getKey(), component);
+
+                continue;
+            }
+
             String s = UNSUPPORTED_FORMAT_PATTERN.matcher(GsonHelper.convertToString(entry.getValue(), entry.getKey())).replaceAll("%$1s");
             p_128110_.accept(entry.getKey(), s);
         }
@@ -90,6 +_,13 @@
 
     public static void inject(Language p_128115_) {
         instance = p_128115_;
+    }
+
+    // Neo: All helpers methods below are injected by Neo to ease modder's usage of Language
+    public Map<String, String> getLanguageData() { return ImmutableMap.of(); }
+
+    public @org.jetbrains.annotations.Nullable net.minecraft.network.chat.Component getComponent(String key) {
+        return null;
     }
 
     public String getOrDefault(String p_128111_) {
