--- a/net/minecraft/world/entity/animal/pig/PigVariant.java
+++ b/net/minecraft/world/entity/animal/pig/PigVariant.java
@@ -6,10 +_,10 @@
 import net.minecraft.core.ClientAsset;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.util.StringRepresentable;
 import net.minecraft.world.entity.variant.ModelAndTexture;
 import net.minecraft.world.entity.variant.PriorityProvider;
@@ -34,7 +_,7 @@
             )
             .apply(i, PigVariant::new)
     );
-    public static final Codec<Holder<PigVariant>> CODEC = RegistryCodecs.holder(Registries.PIG_VARIANT);
+    public static final Codec<Holder<PigVariant>> CODEC = RegistryFixedCodec.create(Registries.PIG_VARIANT);
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<PigVariant>> STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.PIG_VARIANT);
 
     private PigVariant(ModelAndTexture<PigVariant.ModelType> assetInfo, ClientAsset.ResourceTexture babyTexture) {
