--- a/net/minecraft/core/component/BlockTransformer.java
+++ b/net/minecraft/core/component/BlockTransformer.java
@@ -12,7 +_,6 @@
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
@@ -46,11 +_,12 @@
 import org.jspecify.annotations.Nullable;
 
 public record BlockTransformer(List<BlockTransformer.BlockTransformData> transforms) {
-    public static final Codec<BlockTransformer> DIRECT_CODEC = BlockTransformer.BlockTransformData.CODEC
+    public static final Codec<BlockTransformer> CODEC = BlockTransformer.BlockTransformData.CODEC
         .listOf(1, 200)
         .xmap(BlockTransformer::new, BlockTransformer::transforms);
-    public static final Codec<Holder<BlockTransformer>> CODEC = RegistryCodecs.holder(Registries.BLOCK_TRANSFORMER);
-    public static final StreamCodec<RegistryFriendlyByteBuf, Holder<BlockTransformer>> STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.BLOCK_TRANSFORMER);
+    public static final StreamCodec<RegistryFriendlyByteBuf, BlockTransformer> STREAM_CODEC = StreamCodec.composite(
+        BlockTransformer.BlockTransformData.STREAM_CODEC.apply(ByteBufCodecs.list()), BlockTransformer::transforms, BlockTransformer::new
+    );
 
     public InteractionResult transformBlock(UseOnContext context) {
         if (playerHasBlockingItemUseIntent(context)) {
@@ -63,7 +_,7 @@
         for (BlockTransformer.BlockTransformData transformData : this.transforms) {
             Direction clickedFace = context.getClickedFace();
             if (!transformData.disallowedFaces().contains(clickedFace)) {
-                BlockState newBlockState = transformData.blockStateProvider.value().getOptionalState(level, level.getRandom(), pos);
+                BlockState newBlockState = transformData.blockStateProvider.getOptionalState(level, level.getRandom(), pos);
                 if (newBlockState != null) {
                     BlockState updatedShape = transformData.updateFromNeighbors ? Block.updateFromNeighbourShapes(newBlockState, level, pos) : newBlockState;
                     Player player = context.getPlayer();
@@ -79,7 +_,6 @@
                                 lt -> Block.dropFromBlockInteractLootTable(
                                     serverLevel,
                                     (ResourceKey<LootTable>)lt,
-                                    pos,
                                     oldBlockState,
                                     level.getBlockEntity(pos),
                                     itemInHand,
@@ -123,7 +_,7 @@
     }
 
     public record BlockTransformData(
-        Holder<BlockStateProvider> blockStateProvider,
+        BlockStateProvider blockStateProvider,
         Holder<SoundEvent> sound,
         BlockTransformer.TransformParticle particle,
         List<Direction> disallowedFaces,
@@ -181,12 +_,12 @@
             BlockTransformer.BlockTransformData::new
         );
 
-        public static BlockTransformer.BlockTransformData.Builder builder(Holder<BlockStateProvider> targetStateProvider) {
-            return new BlockTransformer.BlockTransformData.Builder(targetStateProvider);
-        }
-
         public static BlockTransformer.BlockTransformData.Builder builder(BlockStateProvider targetStateProvider) {
-            return builder(Holder.direct(targetStateProvider));
+            return new BlockTransformer.BlockTransformData.Builder(targetStateProvider);
+        }
+
+        public static BlockTransformer.BlockTransformData.Builder builder(BlockPredicate predicate, BlockStateProvider stateProvider) {
+            return builder(RuleBasedStateProvider.builder().ifTrueThenProvide(predicate, stateProvider).build());
         }
 
         public static BlockTransformer.BlockTransformData.Builder builder(BlockPredicate predicate, Block block) {
@@ -194,7 +_,7 @@
         }
 
         public static class Builder {
-            private final Holder<BlockStateProvider> targetStateProvider;
+            private final BlockStateProvider targetStateProvider;
             private Holder<SoundEvent> sound = BuiltInRegistries.SOUND_EVENT.wrapAsHolder(SoundEvents.EMPTY);
             private BlockTransformer.TransformParticle particle = BlockTransformer.TransformParticle.NONE;
             private List<Direction> disallowedFaces = List.of();
@@ -205,7 +_,7 @@
             private boolean consumeOnUse = true;
             private int itemDamagePerUse = 1;
 
-            private Builder(Holder<BlockStateProvider> targetStateProvider) {
+            private Builder(BlockStateProvider targetStateProvider) {
                 this.targetStateProvider = targetStateProvider;
             }
 
