--- a/net/minecraft/core/component/DataComponentPatch.java
+++ b/net/minecraft/core/component/DataComponentPatch.java
@@ -61,7 +_,7 @@
         }
     );
     private static final String REMOVED_PREFIX = "!";
-    public final Reference2ObjectMap<DataComponentType<?>, Object> map;
+    final Reference2ObjectMap<DataComponentType<?>, Object> map;
 
     private static StreamCodec<RegistryFriendlyByteBuf, DataComponentPatch> createStreamCodec(DataComponentPatch.CodecGetter codecGetter) {
         return new StreamCodec<RegistryFriendlyByteBuf, DataComponentPatch>() {
@@ -130,7 +_,13 @@
             }
 
             private <T> void encodeComponent(RegistryFriendlyByteBuf output, DataComponentType<T> type, Object value) {
-                codecGetter.apply(type).encode(output, (T)value);
+                // Paper start - codec errors of random anonymous classes are useless
+                try {
+                    codecGetter.apply(type).encode(output, (T)value);
+                } catch (final Exception e) {
+                    throw new RuntimeException("Error encoding component " + type, e);
+                }
+                // Paper end - codec errors of random anonymous classes are useless
             }
         };
     }
@@ -143,6 +_,23 @@
         return new DataComponentPatch.Builder();
     }
 
+    public boolean isPatched(DataComponentType<?> type) {
+        return this.map.containsKey(type);
+    }
+
+    /// {@return the patch stored for the given component type, null if no patch is stored}
+    public <T> @Nullable T getPatch(DataComponentType<? extends T> type) {
+        Object value = this.map.get(type);
+        if (value == null) {
+            throw new IllegalStateException("DataComponentType " + type + " is not patched");
+        }
+        return (T) Removed.removedToNull(value);
+    }
+
+    public Set<DataComponentType<?>> keySet(){
+        return this.map.keySet();
+    }
+
     public <T> @Nullable T get(DataComponentGetter prototype, DataComponentType<? extends T> type) {
         return getFromPatchAndPrototype(this.map, prototype, type);
     }
@@ -237,7 +_,44 @@
         private Builder() {
         }
 
+        // CraftBukkit start
+        public void copy(DataComponentPatch orig) {
+            this.map.putAll(orig.map);
+        }
+
+        public void clear(DataComponentType<?> type) {
+            this.map.remove(type);
+        }
+
+        public boolean isSet(DataComponentType<?> type) {
+            return map.containsKey(type);
+        }
+
+        public boolean isEmpty() {
+            return this.map.isEmpty();
+        }
+
+        @Override
+        public boolean equals(Object object) {
+            if (this == object) {
+                return true;
+            }
+
+            if (object instanceof DataComponentPatch.Builder patch) {
+                return this.map.equals(patch.map);
+            }
+
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return this.map.hashCode();
+        }
+        // CraftBukkit end
+
         public <T> DataComponentPatch.Builder set(DataComponentType<T> type, T value) {
+            net.neoforged.neoforge.common.CommonHooks.validateComponent(value);
             this.map.put(type, value);
             return this;
         }
