--- a/net/minecraft/core/component/PatchedDataComponentMap.java
+++ b/net/minecraft/core/component/PatchedDataComponentMap.java
@@ -8,6 +_,7 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.Set;
 import java.util.Map.Entry;
 import java.util.stream.Collectors;
@@ -15,14 +_,14 @@
 
 public final class PatchedDataComponentMap implements DataComponentMap {
     private final DataComponentMap prototype;
-    private Reference2ObjectMap<DataComponentType<?>, Object> patch;
+    private Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch;
     private boolean copyOnWrite;
 
     public PatchedDataComponentMap(DataComponentMap prototype) {
         this(prototype, Reference2ObjectMaps.emptyMap(), true);
     }
 
-    private PatchedDataComponentMap(DataComponentMap prototype, Reference2ObjectMap<DataComponentType<?>, Object> patch, boolean copyOnWrite) {
+    private PatchedDataComponentMap(DataComponentMap prototype, Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch, boolean copyOnWrite) {
         this.prototype = prototype;
         this.patch = patch;
         this.copyOnWrite = copyOnWrite;
@@ -38,15 +_,15 @@
         return map;
     }
 
-    private static boolean isPatchSanitized(DataComponentMap prototype, Reference2ObjectMap<DataComponentType<?>, Object> patch) {
-        for (Entry<DataComponentType<?>, Object> entry : Reference2ObjectMaps.fastIterable(patch)) {
+    private static boolean isPatchSanitized(DataComponentMap prototype, Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch) {
+        for (Entry<DataComponentType<?>, Optional<?>> entry : Reference2ObjectMaps.fastIterable(patch)) {
             Object defaultValue = prototype.get(entry.getKey());
-            Object value = entry.getValue();
-            if (value.equals(defaultValue)) {
+            Optional<?> value = entry.getValue();
+            if (value.isPresent() && value.get().equals(defaultValue)) {
                 return false;
             }
 
-            if (Removed.isRemoved(value) && defaultValue == null) {
+            if (value.isEmpty() && defaultValue == null) {
                 return false;
             }
         }
@@ -64,53 +_,54 @@
     }
 
     public <T> @Nullable T set(DataComponentType<T> type, @Nullable T value) {
+        net.neoforged.neoforge.common.CommonHooks.validateComponent(value);
         this.ensureMapOwnership();
         T defaultValue = this.prototype.get(type);
-        Object lastValue;
+        Optional<T> lastValue;
         if (Objects.equals(value, defaultValue)) {
-            lastValue = this.patch.remove(type);
+            lastValue = (Optional<T>)this.patch.remove(type);
         } else {
-            lastValue = this.patch.put(type, Removed.nullToRemoved(value));
+            lastValue = (Optional<T>)this.patch.put(type, Optional.ofNullable(value));
         }
 
-        return lastValue != null ? Objects.requireNonNullElse((T)Removed.removedToNull(lastValue), defaultValue) : defaultValue;
+        return lastValue != null ? lastValue.orElse(defaultValue) : defaultValue;
     }
 
     public <T> @Nullable T set(TypedDataComponent<T> value) {
         return this.set(value.type(), value.value());
     }
 
-    public <T> T remove(DataComponentType<? extends T> type) {
+    public <T> @Nullable T remove(DataComponentType<? extends T> type) {
         this.ensureMapOwnership();
         T defaultValue = this.prototype.get(type);
-        Object lastValue;
+        Optional<? extends T> lastValue;
         if (defaultValue != null) {
-            lastValue = this.patch.put(type, Removed.INSTANCE);
+            lastValue = (Optional<? extends T>)this.patch.put(type, Optional.empty());
         } else {
-            lastValue = this.patch.remove(type);
+            lastValue = (Optional<? extends T>)this.patch.remove(type);
         }
 
-        return (T)(lastValue != null ? Removed.removedToNull(lastValue) : defaultValue);
+        return (T)(lastValue != null ? lastValue.orElse(null) : defaultValue);
     }
 
     public void applyPatch(DataComponentPatch patch) {
         this.ensureMapOwnership();
 
-        for (Entry<DataComponentType<?>, Object> entry : Reference2ObjectMaps.fastIterable(patch.map)) {
+        for (Entry<DataComponentType<?>, Optional<?>> entry : Reference2ObjectMaps.fastIterable(patch.map)) {
             this.applyPatch(entry.getKey(), entry.getValue());
         }
     }
 
-    private void applyPatch(DataComponentType<?> type, Object value) {
+    private void applyPatch(DataComponentType<?> type, Optional<?> value) {
         Object defaultValue = this.prototype.get(type);
-        if (Removed.isNotRemoved(value)) {
-            if (value.equals(defaultValue)) {
+        if (value.isPresent()) {
+            if (value.get().equals(defaultValue)) {
                 this.patch.remove(type);
             } else {
                 this.patch.put(type, value);
             }
         } else if (defaultValue != null) {
-            this.patch.put(type, Removed.INSTANCE);
+            this.patch.put(type, Optional.empty());
         } else {
             this.patch.remove(type);
         }
@@ -148,9 +_,9 @@
 
         Set<DataComponentType<?>> components = new ReferenceArraySet<>(this.prototype.keySet());
 
-        for (it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<DataComponentType<?>, Object> entry : Reference2ObjectMaps.fastIterable(this.patch)) {
-            Object value = entry.getValue();
-            if (Removed.isNotRemoved(value)) {
+        for (it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<DataComponentType<?>, Optional<?>> entry : Reference2ObjectMaps.fastIterable(this.patch)) {
+            Optional<?> value = entry.getValue();
+            if (value.isPresent()) {
                 components.add(entry.getKey());
             } else {
                 components.remove(entry.getKey());
@@ -168,10 +_,9 @@
 
         List<TypedDataComponent<?>> components = new ArrayList<>(this.patch.size() + this.prototype.size());
 
-        for (it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<DataComponentType<?>, Object> entry : Reference2ObjectMaps.fastIterable(this.patch)) {
-            Object value = entry.getValue();
-            if (Removed.isNotRemoved(value)) {
-                components.add(TypedDataComponent.createUnchecked(entry.getKey(), value));
+        for (it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<DataComponentType<?>, Optional<?>> entry : Reference2ObjectMaps.fastIterable(this.patch)) {
+            if (entry.getValue().isPresent()) {
+                components.add(TypedDataComponent.createUnchecked(entry.getKey(), entry.getValue().get()));
             }
         }
 
@@ -188,8 +_,8 @@
     public int size() {
         int size = this.prototype.size();
 
-        for (it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<DataComponentType<?>, Object> entry : Reference2ObjectMaps.fastIterable(this.patch)) {
-            boolean inPatch = Removed.isNotRemoved(entry.getValue());
+        for (it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<DataComponentType<?>, Optional<?>> entry : Reference2ObjectMaps.fastIterable(this.patch)) {
+            boolean inPatch = entry.getValue().isPresent();
             boolean inPrototype = this.prototype.has(entry.getKey());
             if (inPatch != inPrototype) {
                 size += inPatch ? 1 : -1;
@@ -199,6 +_,10 @@
         return size;
     }
 
+    public boolean isPatchEmpty() {
+        return this.patch.isEmpty();
+    }
+
     public DataComponentPatch asPatch() {
         if (this.patch.isEmpty()) {
             return DataComponentPatch.EMPTY;
@@ -217,6 +_,13 @@
         return this.patch.isEmpty() ? this.prototype : this.copy();
     }
 
+    /**
+     * {@return true if the contained patch equals the given patch.}
+     */
+    public boolean patchEquals(DataComponentPatch patch) {
+        return this.patch.equals(patch.map);
+    }
+
     @Override
     public boolean equals(Object obj) {
         return this == obj
@@ -226,7 +_,24 @@
 
     @Override
     public int hashCode() {
-        return this.prototype.hashCode() + this.patch.hashCode() * 31;
+        return this.prototype.hashCode() + hashPatch(this.patch) * 31;
+    }
+
+    // Neo: Change implementation of hashCode to reduce collisions.
+    // For a map, hashCode is specified as the sum of the hash codes of its entries.
+    // We do that, but change the entry hash code to 8191^<key hash> * <value hash>,
+    // where <key hash> is the lower bits of the identity hash code of the key.
+    private static int hashPatch(Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch) {
+        int h = 0, n = patch.size();
+        var iterator = it.unimi.dsi.fastutil.objects.Reference2ObjectMaps.fastIterator(patch);
+        while (n-- != 0) {
+            var entry = iterator.next();
+            int exponent = System.identityHashCode(entry.getKey()) & 0xff;
+            // Use 8191 instead of the usual 31, as 31 can produce many collisions with typical integer component ranges (0-255) if the exponent difference is only 1
+            int entryHash = com.google.common.math.IntMath.pow(8191, exponent) * entry.getValue().hashCode();
+            h += entryHash;
+        }
+        return h;
     }
 
     @Override
