--- a/net/minecraft/world/entity/projectile/arrow/Arrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/Arrow.java
@@ -8,7 +_,6 @@
 import net.minecraft.network.syncher.SynchedEntityData;
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntityEvent;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EntityTypes;
 import net.minecraft.world.entity.LivingEntity;
@@ -57,7 +_,7 @@
         this.updateColor();
     }
 
-    private void updateColor() {
+    public void updateColor() {
         PotionContents potionContents = this.getPotionContents();
         this.entityData.set(ID_EFFECT_COLOR, potionContents.equals(PotionContents.EMPTY) ? -1 : potionContents.getColor());
     }
@@ -72,6 +_,16 @@
         entityData.define(ID_EFFECT_COLOR, -1);
     }
 
+    // Paper start - EAR 2
+    @Override
+    public void inactiveTick() {
+        if (this.isInGround()) {
+            this.life++;
+        }
+        super.inactiveTick();
+    }
+    // Paper end
+
     @Override
     public void tick() {
         super.tick();
@@ -117,7 +_,9 @@
         Entity effectSource = this.getEffectSource();
         PotionContents potionContents = this.getPotionContents();
         float durationScale = this.getPotionDurationScale();
-        potionContents.forEachEffect(effect -> mob.addEffect(effect, effectSource), durationScale);
+        potionContents.forEachEffect(effect -> {
+            mob.addEffect(effect, effectSource, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
+        }, durationScale);
     }
 
     @Override
@@ -126,7 +_,7 @@
     }
 
     @Override
-    public void handleEntityEvent(@EntityEvent.Value byte id) {
+    public void handleEntityEvent(byte id) {
         if (id == 0) {
             int colorValue = this.getColor();
             if (colorValue != -1) {
