--- a/net/minecraft/world/entity/projectile/arrow/Arrow.java
+++ b/net/minecraft/world/entity/projectile/arrow/Arrow.java
@@ -71,6 +_,16 @@
         entityData.define(ID_EFFECT_COLOR, -1);
     }
 
+    // Paper start - EAR 2
+    @Override
+    public void inactiveTick() {
+        if (this.isInGround()) {
+            this.life++;
+        }
+        super.inactiveTick();
+    }
+    // Paper end
+
     @Override
     public void tick() {
         super.tick();
@@ -116,7 +_,9 @@
         Entity effectSource = this.getEffectSource();
         PotionContents potionContents = this.getPotionContents();
         float durationScale = this.getPotionDurationScale();
-        potionContents.forEachEffect(effect -> mob.addEffect(effect, effectSource), durationScale);
+        potionContents.forEachEffect(effect -> {
+            mob.addEffect(effect, effectSource, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
+        }, durationScale);
     }
 
     @Override
