--- a/net/minecraft/client/gui/screens/ErrorScreen.java
+++ b/net/minecraft/client/gui/screens/ErrorScreen.java
@@ -1,9 +_,7 @@
 package net.minecraft.client.gui.screens;
 
 import net.minecraft.client.gui.GuiGraphicsExtractor;
-import net.minecraft.client.gui.TextAlignment;
 import net.minecraft.client.gui.components.Button;
-import net.minecraft.client.gui.components.MultiLineLabel;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.neoforged.api.distmarker.Dist;
@@ -11,18 +_,16 @@
 
 @OnlyIn(Dist.CLIENT)
 public class ErrorScreen extends Screen {
-    private final Component messageText;
-    private MultiLineLabel message;
+    private final Component message;
 
-    public ErrorScreen(Component title, Component messageText) {
+    public ErrorScreen(Component title, Component message) {
         super(title);
-        this.messageText = messageText;
+        this.message = message;
     }
 
     @Override
     protected void init() {
         super.init();
-        this.message = MultiLineLabel.create(this.font, this.messageText, this.width - 50);
         this.addRenderableWidget(
             Button.builder(CommonComponents.GUI_CANCEL, button -> this.minecraft.gui.setScreen(null)).bounds(this.width / 2 - 100, 140, 200, 20).build()
         );
@@ -32,7 +_,7 @@
     public void extractRenderState(GuiGraphicsExtractor graphics, int mouseX, int mouseY, float a) {
         super.extractRenderState(graphics, mouseX, mouseY, a);
         graphics.centeredText(this.font, this.title, this.width / 2, 90, -1);
-        this.message.visitLines(TextAlignment.CENTER, this.width / 2, 110, 9, graphics.textRenderer());
+        graphics.centeredText(this.font, this.message, this.width / 2, 110, -1);
     }
 
     @Override
