--- a/net/minecraft/client/gui/screens/LevelLoadingScreen.java
+++ b/net/minecraft/client/gui/screens/LevelLoadingScreen.java
@@ -43,7 +_,9 @@
         map.put(ChunkStatus.STRUCTURE_STARTS, 10066329);
         map.put(ChunkStatus.STRUCTURE_REFERENCES, 6250897);
         map.put(ChunkStatus.BIOMES, 8434258);
-        map.put(ChunkStatus.TERRAIN, 3159410);
+        map.put(ChunkStatus.NOISE, 13750737);
+        map.put(ChunkStatus.SURFACE, 7497737);
+        map.put(ChunkStatus.CARVERS, 3159410);
         map.put(ChunkStatus.FEATURES, 2213376);
         map.put(ChunkStatus.INITIALIZE_LIGHT, 13421772);
         map.put(ChunkStatus.LIGHT, 16769184);
