--- a/net/minecraft/client/gui/screens/LoadingOverlay.java
+++ b/net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -46,7 +_,7 @@
     private final Consumer<Optional<Throwable>> onFinish;
     private final boolean fadeIn;
     private float currentProgress;
-    protected long fadeOutStart = -1L;
+    private long fadeOutStart = -1L;
     private long fadeInStart = -1L;
 
     public LoadingOverlay(Minecraft minecraft, ReloadInstance reload, Consumer<Optional<Throwable>> onFinish, boolean fadeIn) {
