--- a/net/minecraft/client/gui/screens/LoadingOverlay.java
+++ b/net/minecraft/client/gui/screens/LoadingOverlay.java
@@ -194,7 +_,7 @@
 
         @Override
         public TextureContents loadContents(ResourceManager resourceManager) throws IOException {
-            ResourceProvider vanillaProvider = Minecraft.getInstance().getVanillaPackResources().asResourceManager();
+            ResourceProvider vanillaProvider = Minecraft.getInstance().getVanillaPackResources().asProvider();
 
             try (InputStream resource = vanillaProvider.open(LoadingOverlay.MOJANG_STUDIOS_LOGO_LOCATION)) {
                 return new TextureContents(NativeImage.read(resource), new TextureMetadataSection(true, true, MipmapStrategy.MEAN, 0.0F));
