--- a/net/minecraft/client/gui/screens/MenuScreens.java
+++ b/net/minecraft/client/gui/screens/MenuScreens.java
@@ -103,7 +_,7 @@
         register(MenuType.STONECUTTER, StonecutterScreen::new);
     }
 
-    public interface ScreenConstructor<T extends AbstractContainerMenu, U extends Screen & MenuAccess<T>> {
+    private interface ScreenConstructor<T extends AbstractContainerMenu, U extends Screen & MenuAccess<T>> {
         default void fromPacket(Component title, MenuType<T> type, Minecraft minecraft, int containerId) {
             U screen = this.create(type.create(containerId, minecraft.player.getInventory()), minecraft.player.getInventory(), title);
             minecraft.player.containerMenu = screen.getMenu();
