--- a/net/minecraft/client/gui/screens/Screen.java
+++ b/net/minecraft/client/gui/screens/Screen.java
@@ -68,7 +_,7 @@
     private boolean initialized;
     public int width;
     public int height;
-    public final List<Renderable> renderables = Lists.newArrayList();
+    private final List<Renderable> renderables = Lists.newArrayList();
     protected final Font font;
     private static final long NARRATE_SUPPRESS_AFTER_INIT_TIME = TimeUnit.SECONDS.toMillis(2L);
     private static final long NARRATE_DELAY_NARRATOR_ENABLED = NARRATE_SUPPRESS_AFTER_INIT_TIME;
