--- a/net/minecraft/client/gui/screens/TitleScreen.java
+++ b/net/minecraft/client/gui/screens/TitleScreen.java
@@ -142,7 +_,7 @@
         );
         accessibility.setPosition(this.getHorizontalPosition(++currentButton, 3, 20), topPos);
         Button.Builder var10001 = Button.builder(
-            Component.translatable("menu.options"), var1x -> this.minecraft.gui.setScreen(new OptionsScreen(this, this.minecraft.options))
+            Component.translatable("menu.options"), var1x -> this.minecraft.gui.setScreen(new OptionsScreen(this, this.minecraft.options, false))
         );
         int var10002 = this.width / 2 - 100;
         topPos += 24;
