--- a/net/minecraft/data/BlockFamilies.java
+++ b/net/minecraft/data/BlockFamilies.java
@@ -13,7 +_,6 @@
 public class BlockFamilies {
     private static final Map<Block, BlockFamily> MAP = Maps.newHashMap();
     private static final String RECIPE_GROUP_PREFIX_WOODEN = "wooden";
-    private static final String RECIPE_GROUP_PREFIX_WOOLEN = "woolen";
     private static final String RECIPE_UNLOCKED_BY_HAS_PLANKS = "has_planks";
     public static final BlockFamily ACACIA_PLANKS = familyBuilder(Blocks.ACACIA_PLANKS)
         .log(Blocks.ACACIA_LOG)
@@ -563,16 +_,6 @@
                 .carpet(Blocks.CARPET.pick(color))
                 .stairs(Blocks.WOOL_STAIRS.pick(color))
                 .slab(Blocks.WOOL_SLAB.pick(color))
-                .recipeGroupPrefix("woolen")
-                .getFamily()
-        );
-    public static final ColorCollection<BlockFamily> CONCRETE = ColorCollection.VALUES
-        .map(
-            color -> familyBuilder(Blocks.CONCRETE.pick(color))
-                .stairs(Blocks.CONCRETE_STAIRS.pick(color))
-                .slab(Blocks.CONCRETE_SLAB.pick(color))
-                .recipeGroupPrefix("concrete")
-                .generateStonecutterRecipe()
                 .getFamily()
         );
 
