--- a/net/minecraft/world/entity/monster/EntityPhantom.java
+++ b/net/minecraft/world/entity/monster/EntityPhantom.java
@@ -164,7 +164,7 @@
     @Override
     protected void readAdditionalSaveData(ValueInput valueinput) {
         super.readAdditionalSaveData(valueinput);
-        this.anchorPoint = (BlockPosition) valueinput.read("anchor_pos", BlockPosition.CODEC).orElse((Object) null);
+        this.anchorPoint = (BlockPosition) valueinput.read("anchor_pos", BlockPosition.CODEC).orElse(null); // CraftBukkit - decompile error
         this.setPhantomSize(valueinput.getIntOr("size", 0));
     }
 
@@ -534,11 +534,11 @@
                 List<EntityHuman> list = worldserver.getNearbyPlayers(this.attackTargeting, EntityPhantom.this, EntityPhantom.this.getBoundingBox().inflate(16.0D, 64.0D, 16.0D));
 
                 if (!list.isEmpty()) {
-                    list.sort(Comparator.comparing(Entity::getY).reversed());
+                    list.sort(Comparator.comparing((Entity e) -> { return e.getY(); }).reversed()); // CraftBukkit - decompile error
 
                     for (EntityHuman entityhuman : list) {
                         if (EntityPhantom.this.canAttack(worldserver, entityhuman, PathfinderTargetCondition.DEFAULT)) {
-                            EntityPhantom.this.setTarget(entityhuman);
+                            EntityPhantom.this.setTarget(entityhuman, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER, true); // CraftBukkit - reason
                             return true;
                         }
                     }
