--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -109,7 +109,7 @@
     @Override
     protected void readAdditionalSaveData(ValueInput input) {
         super.readAdditionalSaveData(input);
-        this.boundOrigin = (BlockPos) input.read("bound_pos", BlockPos.CODEC).orElse((Object) null);
+        this.boundOrigin = (BlockPos) input.read("bound_pos", BlockPos.CODEC).orElse(null); // CraftBukkit - decompile error
         input.getInt("life_ticks").ifPresentOrElse(this::setLimitedLife, () -> {
             this.hasLimitedLife = false;
         });
@@ -383,7 +383,7 @@
         public void start() {
             Mob mob = Vex.this.getOwner();
 
-            Vex.this.setTarget(mob != null ? mob.getTarget() : null);
+            Vex.this.setTarget(mob != null ? mob.getTarget() : null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit
             super.start();
         }
     }
