--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -108,7 +108,7 @@
     @Override
     protected void readAdditionalSaveData(ValueInput valueinput) {
         super.readAdditionalSaveData(valueinput);
-        this.boundOrigin = (BlockPos) valueinput.read("bound_pos", BlockPos.CODEC).orElse((Object) null);
+        this.boundOrigin = (BlockPos) valueinput.read("bound_pos", BlockPos.CODEC).orElse(null); // CraftBukkit - decompile error
         valueinput.getInt("life_ticks").ifPresentOrElse(this::setLimitedLife, () -> {
             this.hasLimitedLife = false;
         });
@@ -375,7 +375,7 @@
         public void start() {
             Mob mob = Vex.this.getOwner();
 
-            Vex.this.setTarget(mob != null ? mob.getTarget() : null);
+            Vex.this.setTarget(mob != null ? mob.getTarget() : null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit
             super.start();
         }
     }
