--- a/net/minecraft/world/level/block/AbstractBedBlock.java
+++ b/net/minecraft/world/level/block/AbstractBedBlock.java
@@ -2,7 +_,6 @@
 
 import java.util.List;
 import java.util.Optional;
-import java.util.OptionalDouble;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.network.chat.Component;
@@ -36,7 +_,6 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
-import net.minecraft.world.phys.shapes.VoxelShape;
 import org.apache.commons.lang3.ArrayUtils;
 import org.jspecify.annotations.Nullable;
 
@@ -68,13 +_,8 @@
         return Stats.SLEEP_IN_BED;
     }
 
-    public OptionalDouble getSleepHeight(BlockState state, Level level, BlockPos pos) {
-        if (!state.is(this)) {
-            return OptionalDouble.empty();
-        }
-
-        VoxelShape shape = state.getShape(level, pos);
-        return shape.isEmpty() ? OptionalDouble.empty() : OptionalDouble.of(shape.max(Direction.Axis.Y));
+    public double getSleepHeight(BlockState state, Level level, BlockPos pos) {
+        return state.getShape(level, pos).max(Direction.Axis.Y);
     }
 
     @Override
