--- a/net/minecraft/world/level/block/AbstractCauldronBlock.java
+++ b/net/minecraft/world/level/block/AbstractCauldronBlock.java
@@ -53,7 +_,7 @@
         ItemStack itemStack, BlockState state, Level level, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hitResult
     ) {
         CauldronInteraction behavior = this.interactions.get(itemStack);
-        return behavior.interact(state, level, pos, player, hand, itemStack);
+        return behavior.interact(state, level, pos, player, hand, itemStack, hitResult.getDirection()); // Paper - pass hit direction
     }
 
     @Override
@@ -94,5 +_,23 @@
     }
 
     protected void receiveStalactiteDrip(BlockState state, Level level, BlockPos pos, Fluid fluid) {
+    }
+
+    @Override
+    public void onPlace(BlockState state, Level level, BlockPos pos, BlockState oldState, boolean movedByPiston) {
+        super.onPlace(state, level, pos, oldState, movedByPiston);
+        // Neo: Invalidate cauldron capabilities when a cauldron is added
+        if (net.neoforged.neoforge.fluids.CauldronFluidContent.getForBlock(oldState.getBlock()) == null) {
+            level.invalidateCapabilities(pos);
+        }
+    }
+
+    @Override
+    protected void affectNeighborsAfterRemoval(BlockState state, ServerLevel level, BlockPos pos, boolean movedByPiston) {
+        super.affectNeighborsAfterRemoval(state, level, pos, movedByPiston);
+        // Neo: Invalidate cauldron capabilities when a cauldron is removed
+        if (net.neoforged.neoforge.fluids.CauldronFluidContent.getForBlock(level.getBlockState(pos).getBlock()) == null) {
+            level.invalidateCapabilities(pos);
+        }
     }
 }
