--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -8,7 +_,6 @@
 import net.minecraft.tags.BlockTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.entity.player.Player;
-import net.minecraft.world.item.SwordItem;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
@@ -26,6 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class BambooStalkBlock extends Block implements BonemealableBlock {
     public static final MapCodec<BambooStalkBlock> CODEC = simpleCodec(BambooStalkBlock::new);
@@ -97,7 +_,8 @@
             return null;
         } else {
             BlockState blockstate = p_261764_.getLevel().getBlockState(p_261764_.getClickedPos().below());
-            if (blockstate.is(BlockTags.BAMBOO_PLANTABLE_ON)) {
+            net.neoforged.neoforge.common.util.TriState soilDecision = blockstate.canSustainPlant(p_261764_.getLevel(), p_261764_.getClickedPos().below(), net.minecraft.core.Direction.UP, this.defaultBlockState());
+            if (soilDecision.isDefault() ? blockstate.is(BlockTags.BAMBOO_PLANTABLE_ON) : soilDecision.isTrue()) {
                 if (blockstate.is(Blocks.BAMBOO_SAPLING)) {
                     return this.defaultBlockState().setValue(AGE, Integer.valueOf(0));
                 } else if (blockstate.is(Blocks.BAMBOO)) {
@@ -130,10 +_,11 @@
     @Override
     protected void randomTick(BlockState p_261931_, ServerLevel p_261751_, BlockPos p_261616_, RandomSource p_261766_) {
         if (p_261931_.getValue(STAGE) == 0) {
-            if (p_261766_.nextInt(3) == 0 && p_261751_.isEmptyBlock(p_261616_.above()) && p_261751_.getRawBrightness(p_261616_.above(), 0) >= 9) {
+            if (p_261751_.isEmptyBlock(p_261616_.above()) && p_261751_.getRawBrightness(p_261616_.above(), 0) >= 9) {
                 int i = this.getHeightBelowUpToMax(p_261751_, p_261616_) + 1;
-                if (i < 16) {
+                if (i < p_261751_.paperConfig().maxGrowthHeight.bamboo.max && net.neoforged.neoforge.common.CommonHooks.canCropGrow(p_261751_, p_261616_, p_261931_, p_261766_.nextFloat() < (p_261751_.spigotConfig.bambooModifier / (100.0f * 3)))) { // Spigot
                     this.growBamboo(p_261931_, p_261751_, p_261616_, p_261766_, i);
+                    net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(p_261751_, p_261616_, p_261931_);
                 }
             }
         }
@@ -141,6 +_,8 @@
 
     @Override
     protected boolean canSurvive(BlockState p_261860_, LevelReader p_262154_, BlockPos p_261493_) {
+        net.neoforged.neoforge.common.util.TriState soilDecision = p_262154_.getBlockState(p_261493_.below()).canSustainPlant(p_262154_, p_261493_.below(), Direction.UP, p_261860_);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return p_262154_.getBlockState(p_261493_.below()).is(BlockTags.BAMBOO_PLANTABLE_ON);
     }
 
@@ -163,7 +_,7 @@
     public boolean isValidBonemealTarget(LevelReader p_262065_, BlockPos p_262033_, BlockState p_261700_) {
         int i = this.getHeightAboveUpToMax(p_262065_, p_262033_);
         int j = this.getHeightBelowUpToMax(p_262065_, p_262033_);
-        return i + j + 1 < 16 && p_262065_.getBlockState(p_262033_.above(i)).getValue(STAGE) != 1;
+        return i + j + 1 < ((Level) p_262065_).paperConfig().maxGrowthHeight.bamboo.max && p_262065_.getBlockState(p_262033_.above(i)).getValue(STAGE) != 1;
     }
 
     @Override
@@ -181,7 +_,7 @@
         for (int i1 = 0; i1 < l; i1++) {
             BlockPos blockpos = p_261955_.above(i);
             BlockState blockstate = p_261845_.getBlockState(blockpos);
-            if (k >= 16 || blockstate.getValue(STAGE) == 1 || !p_261845_.isEmptyBlock(blockpos.above())) {
+            if (k >= p_261845_.paperConfig().maxGrowthHeight.bamboo.max || blockstate.getValue(STAGE) == 1 || !p_261845_.isEmptyBlock(blockpos.above())) {
                 return;
             }
 
@@ -193,7 +_,7 @@
 
     @Override
     protected float getDestroyProgress(BlockState p_261691_, Player p_262171_, BlockGetter p_261621_, BlockPos p_261500_) {
-        return p_262171_.getMainHandItem().getItem() instanceof SwordItem ? 1.0F : super.getDestroyProgress(p_261691_, p_262171_, p_261621_, p_261500_);
+        return p_262171_.getMainHandItem().canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SWORD_DIG) ? 1.0F : super.getDestroyProgress(p_261691_, p_262171_, p_261621_, p_261500_);
     }
 
     protected void growBamboo(BlockState p_261855_, Level p_262076_, BlockPos p_262109_, RandomSource p_261633_, int p_261759_) {
@@ -201,29 +_,42 @@
         BlockPos blockpos = p_262109_.below(2);
         BlockState blockstate1 = p_262076_.getBlockState(blockpos);
         BambooLeaves bambooleaves = BambooLeaves.NONE;
+        boolean shouldUpdateOthers = false; // CraftBukkit
+
         if (p_261759_ >= 1) {
             if (!blockstate.is(Blocks.BAMBOO) || blockstate.getValue(LEAVES) == BambooLeaves.NONE) {
                 bambooleaves = BambooLeaves.SMALL;
             } else if (blockstate.is(Blocks.BAMBOO) && blockstate.getValue(LEAVES) != BambooLeaves.NONE) {
                 bambooleaves = BambooLeaves.LARGE;
                 if (blockstate1.is(Blocks.BAMBOO)) {
-                    p_262076_.setBlock(p_262109_.below(), blockstate.setValue(LEAVES, BambooLeaves.SMALL), 3);
-                    p_262076_.setBlock(blockpos, blockstate1.setValue(LEAVES, BambooLeaves.NONE), 3);
+                    // CraftBukkit start - moved down
+                    // p_262076_.setBlock(p_262109_.below(), blockstate.setValue(LEAVES, BambooLeaves.SMALL), 3);
+                    // p_262076_.setBlock(blockpos, blockstate1.setValue(LEAVES, BambooLeaves.NONE), 3);
+                    shouldUpdateOthers = true;
+                    // CraftBukkit end
                 }
             }
         }
 
         int i = p_261855_.getValue(AGE) != 1 && !blockstate1.is(Blocks.BAMBOO) ? 0 : 1;
-        int j = (p_261759_ < 11 || !(p_261633_.nextFloat() < 0.25F)) && p_261759_ != 15 ? 0 : 1;
-        p_262076_.setBlock(
-            p_262109_.above(), this.defaultBlockState().setValue(AGE, Integer.valueOf(i)).setValue(LEAVES, bambooleaves).setValue(STAGE, Integer.valueOf(j)), 3
-        );
+        int j = (p_261759_ < p_262076_.paperConfig().maxGrowthHeight.bamboo.min || !(p_261633_.nextFloat() < 0.25F)) && p_261759_ != (p_262076_.paperConfig().maxGrowthHeight.bamboo.max - 1) ? 0 : 1;
+
+        // CraftBukkit start
+        if (CraftEventFactory.handleBlockSpreadEvent(
+                p_262076_, p_262109_, p_262109_.above(),
+                this.defaultBlockState().setValue(AGE, i).setValue(LEAVES, bambooleaves).setValue(STAGE, j), 3)) {
+            if (shouldUpdateOthers) {
+                p_262076_.setBlock(p_262109_.below(), blockstate.setValue(LEAVES, BambooLeaves.SMALL), 3);
+                p_262076_.setBlock(blockpos, blockstate1.setValue(LEAVES, BambooLeaves.NONE), 3);
+            }
+        }
+        // CraftBukkit end
     }
 
     protected int getHeightAboveUpToMax(BlockGetter p_261541_, BlockPos p_261593_) {
         int i = 0;
 
-        while (i < 16 && p_261541_.getBlockState(p_261593_.above(i + 1)).is(Blocks.BAMBOO)) {
+        while (i < ((Level) p_261541_).paperConfig().maxGrowthHeight.bamboo.max && p_261541_.getBlockState(p_261593_.above(i + 1)).is(Blocks.BAMBOO)) {
             i++;
         }
 
@@ -233,7 +_,7 @@
     protected int getHeightBelowUpToMax(BlockGetter p_261927_, BlockPos p_261481_) {
         int i = 0;
 
-        while (i < 16 && p_261927_.getBlockState(p_261481_.below(i + 1)).is(Blocks.BAMBOO)) {
+        while (i < ((Level) p_261927_).paperConfig().maxGrowthHeight.bamboo.max && p_261927_.getBlockState(p_261481_.below(i + 1)).is(Blocks.BAMBOO)) {
             i++;
         }
 
