--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -80,7 +_,8 @@
         }
 
         BlockState belowState = context.getLevel().getBlockState(context.getClickedPos().below());
-        if (belowState.is(BlockTags.SUPPORTS_BAMBOO)) {
+        var soilDecision = belowState.canSustainPlant(context.getLevel(), context.getClickedPos().below(), net.minecraft.core.Direction.UP, this.defaultBlockState());
+        if (soilDecision.isDefault() ? belowState.is(BlockTags.SUPPORTS_BAMBOO) : soilDecision.isTrue()) {
             if (belowState.is(Blocks.BAMBOO_SAPLING)) {
                 return this.defaultBlockState().setValue(AGE, 0);
             } else if (belowState.is(Blocks.BAMBOO)) {
@@ -112,10 +_,11 @@
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (state.getValue(STAGE) == 0) {
-            if (random.nextInt(3) == 0 && level.isEmptyBlock(pos.above()) && level.getRawBrightness(pos.above(), 0) >= 9) {
+            if (level.isEmptyBlock(pos.above()) && level.getRawBrightness(pos.above(), 0) >= 9) {
                 int height = this.getHeightBelowUpToMax(level, pos) + 1;
-                if (height < 16) {
+                if (height < level.paperConfig().maxGrowthHeight.bamboo.max && net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, random.nextFloat() < (level.spigotConfig.bambooModifier / (100.0F * 3)) )) {
                     this.growBamboo(state, level, pos, random, height);
+                    net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos, state);
                 }
             }
         }
@@ -123,6 +_,8 @@
 
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
+        var soilDecision = level.getBlockState(pos.below()).canSustainPlant(level, pos.below(), Direction.UP, state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return level.getBlockState(pos.below()).is(BlockTags.SUPPORTS_BAMBOO);
     }
 
@@ -147,23 +_,23 @@
     }
 
     @Override
-    public boolean isValidBonemealTarget(LevelReader level, BlockPos pos, BlockState state, BonemealSource source) {
+    public boolean isValidBonemealTarget(LevelReader level, BlockPos pos, BlockState state) {
         int heightAbove = this.getHeightAboveUpToMax(level, pos);
         int heightBelow = this.getHeightBelowUpToMax(level, pos);
         BlockPos growthPos = pos.above(heightAbove + 1);
-        return heightAbove + heightBelow + 1 < 16
+        return heightAbove + heightBelow + 1 < ((Level) level).paperConfig().maxGrowthHeight.bamboo.max // Paper - Configurable cactus/bamboo/reed growth height
             && level.getBlockState(pos.above(heightAbove)).getValue(STAGE) != 1
             && level.isInsideBuildHeight(growthPos)
             && level.isEmptyBlock(growthPos);
     }
 
     @Override
-    public boolean isBonemealSuccess(Level level, RandomSource random, BlockPos pos, BlockState state, BonemealSource source) {
+    public boolean isBonemealSuccess(Level level, RandomSource random, BlockPos pos, BlockState state) {
         return true;
     }
 
     @Override
-    public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state, BonemealSource source) {
+    public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state) {
         int heightAbove = this.getHeightAboveUpToMax(level, pos);
         int heightBelow = this.getHeightBelowUpToMax(level, pos);
         int totalHeight = heightAbove + heightBelow + 1;
@@ -173,7 +_,7 @@
             BlockPos topPos = pos.above(heightAbove);
             BlockState topState = level.getBlockState(topPos);
             BlockPos growthPos = topPos.above();
-            if (totalHeight >= 16 || topState.getValue(STAGE) == 1 || !level.isEmptyBlock(growthPos) || level.isOutsideBuildHeight(growthPos)) {
+            if (totalHeight >= level.paperConfig().maxGrowthHeight.bamboo.max || !topState.is(Blocks.BAMBOO) || topState.getValue(STAGE) == 1 || !level.isEmptyBlock(growthPos) || level.isOutsideBuildHeight(growthPos)) { // CraftBukkit - If the BlockSpreadEvent was cancelled, we have no bamboo here // Paper - Configurable cactus/bamboo/reed growth height
                 return;
             }
 
@@ -188,27 +_,39 @@
         BlockPos twoBelowPos = pos.below(2);
         BlockState twoBelowState = level.getBlockState(twoBelowPos);
         BambooLeaves leaves = BambooLeaves.NONE;
+        boolean shouldUpdateOthers = false; // CraftBukkit
         if (height >= 1) {
             if (!belowState.is(Blocks.BAMBOO) || belowState.getValue(LEAVES) == BambooLeaves.NONE) {
                 leaves = BambooLeaves.SMALL;
             } else if (belowState.is(Blocks.BAMBOO) && belowState.getValue(LEAVES) != BambooLeaves.NONE) {
                 leaves = BambooLeaves.LARGE;
                 if (twoBelowState.is(Blocks.BAMBOO)) {
-                    level.setBlockAndUpdate(pos.below(), belowState.setValue(LEAVES, BambooLeaves.SMALL));
-                    level.setBlockAndUpdate(twoBelowPos, twoBelowState.setValue(LEAVES, BambooLeaves.NONE));
+                    // CraftBukkit start - moved down
+                    // level.setBlock(pos.below(), belowState.setValue(LEAVES, BambooLeaves.SMALL), Block.UPDATE_ALL);
+                    // level.setBlock(twoBelowPos, twoBelowState.setValue(LEAVES, BambooLeaves.NONE), Block.UPDATE_ALL);
+                    shouldUpdateOthers = true;
+                    // CraftBukkit end
                 }
             }
         }
 
         int age = state.getValue(AGE) != 1 && !twoBelowState.is(Blocks.BAMBOO) ? 0 : 1;
         int stage = (height < 11 || !(random.nextFloat() < 0.25F)) && height != 15 ? 0 : 1;
-        level.setBlockAndUpdate(pos.above(), this.defaultBlockState().setValue(AGE, age).setValue(LEAVES, leaves).setValue(STAGE, stage));
+        // level.setBlock(pos.above(), this.defaultBlockState().setValue(AGE, age).setValue(LEAVES, leaves).setValue(STAGE, stage), Block.UPDATE_ALL);
+        // CraftBukkit start
+        if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, pos.above(), this.defaultBlockState().setValue(AGE, age).setValue(LEAVES, leaves).setValue(STAGE, stage), Block.UPDATE_ALL)) {
+            if (shouldUpdateOthers) {
+                level.setBlock(pos.below(), belowState.setValue(LEAVES, BambooLeaves.SMALL), Block.UPDATE_ALL);
+                level.setBlock(twoBelowPos, twoBelowState.setValue(LEAVES, BambooLeaves.NONE), Block.UPDATE_ALL);
+            }
+        }
+        // CraftBukkit end
     }
 
     protected int getHeightAboveUpToMax(BlockGetter level, BlockPos pos) {
         int height = 0;
 
-        while (height < 16 && level.getBlockState(pos.above(height + 1)).is(Blocks.BAMBOO)) {
+        while (height < ((Level) level).paperConfig().maxGrowthHeight.bamboo.max && level.getBlockState(pos.above(height + 1)).is(Blocks.BAMBOO)) {
             height++;
         }
 
@@ -218,7 +_,7 @@
     protected int getHeightBelowUpToMax(BlockGetter level, BlockPos pos) {
         int height = 0;
 
-        while (height < 16 && level.getBlockState(pos.below(height + 1)).is(Blocks.BAMBOO)) {
+        while (height < ((Level) level).paperConfig().maxGrowthHeight.bamboo.max && level.getBlockState(pos.below(height + 1)).is(Blocks.BAMBOO)) {
             height++;
         }
 
