--- a/net/minecraft/world/level/block/BarrelBlock.java
+++ b/net/minecraft/world/level/block/BarrelBlock.java
@@ -45,8 +_,7 @@
             return InteractionResult.SUCCESS;
         } else {
             BlockEntity blockentity = p_49070_.getBlockEntity(p_49071_);
-            if (blockentity instanceof BarrelBlockEntity) {
-                p_49072_.openMenu((BarrelBlockEntity)blockentity);
+            if (blockentity instanceof BarrelBlockEntity && p_49072_.openMenu((BarrelBlockEntity)blockentity).isPresent()) { // Paper - Fix InventoryOpenEvent cancellation
                 p_49072_.awardStat(Stats.OPEN_BARREL);
                 PiglinAi.angerNearbyPiglins(p_49072_, true);
             }
