--- a/net/minecraft/world/level/block/BaseEntityBlock.java
+++ b/net/minecraft/world/level/block/BaseEntityBlock.java
@@ -33,7 +_,7 @@
 
     @Nullable
     @Override
-    protected MenuProvider getMenuProvider(BlockState p_49234_, Level p_49235_, BlockPos p_49236_) {
+    public MenuProvider getMenuProvider(BlockState p_49234_, Level p_49235_, BlockPos p_49236_) {
         BlockEntity blockentity = p_49235_.getBlockEntity(p_49236_);
         return blockentity instanceof MenuProvider ? (MenuProvider)blockentity : null;
     }
