--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -11,6 +_,7 @@
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.context.BlockPlaceContext;
+import net.minecraft.world.item.context.UseOnContext;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.state.BlockBehaviour;
@@ -18,6 +_,8 @@
 import net.minecraft.world.level.portal.PortalShape;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public abstract class BaseFireBlock extends Block {
     private static final int SECONDS_ON_FIRE = 8;
@@ -127,10 +_,22 @@
 
     @Override
     protected void entityInside(BlockState p_49260_, Level p_49261_, BlockPos p_49262_, Entity p_49263_) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(p_49263_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_49261_, p_49262_)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (!p_49263_.fireImmune()) {
             p_49263_.setRemainingFireTicks(p_49263_.getRemainingFireTicks() + 1);
             if (p_49263_.getRemainingFireTicks() == 0) {
-                p_49263_.igniteForSeconds(8.0F);
+                // CraftBukkit start
+                org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustByBlockEvent(CraftBlock.at(p_49261_, p_49262_), p_49263_.getBukkitEntity(), 8.0F);
+                p_49261_.getCraftServer().getPluginManager().callEvent(event);
+
+                if (!event.isCancelled()) {
+                    p_49263_.igniteForSeconds(event.getDuration(), false);
+                    // Paper start - fix EntityCombustEvent cancellation
+                } else {
+                    p_49263_.setRemainingFireTicks(p_49263_.getRemainingFireTicks() - 1);
+                    // Paper end - fix EntityCombustEvent cancellation
+                }
+                // CraftBukkit end
             }
         }
 
@@ -138,24 +_,32 @@
         super.entityInside(p_49260_, p_49261_, p_49262_, p_49263_);
     }
 
+
+    public UseOnContext onPlace$context;
     @Override
     protected void onPlace(BlockState p_49279_, Level p_49280_, BlockPos p_49281_, BlockState p_49282_, boolean p_49283_) {
         if (!p_49282_.is(p_49279_.getBlock())) {
             if (inPortalDimension(p_49280_)) {
                 Optional<PortalShape> optional = PortalShape.findEmptyPortalShape(p_49280_, p_49281_, Direction.Axis.X);
+                optional = net.neoforged.neoforge.event.EventHooks.onTrySpawnPortal(p_49280_, p_49281_, optional);
                 if (optional.isPresent()) {
-                    optional.get().createPortalBlocks();
+                    PortalShape portalShape = optional.get();
+                    portalShape.createPortalBlocks$Entity = (onPlace$context == null) ? null : onPlace$context.getPlayer();
+                    portalShape.createPortalBlocks();
                     return;
                 }
             }
 
             if (!p_49279_.canSurvive(p_49280_, p_49281_)) {
-                p_49280_.removeBlock(p_49281_, false);
+                fireExtinguished(p_49280_, p_49281_); // CraftBukkit - fuel block broke
             }
         }
     }
 
     private static boolean inPortalDimension(Level p_49249_) {
+        if (Boolean.parseBoolean("true")) {
+            return p_49249_.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.OVERWORLD || p_49249_.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER; // CraftBukkit - getTypeKey()
+        }
         return p_49249_.dimension() == Level.OVERWORLD || p_49249_.dimension() == Level.NETHER;
     }
 
@@ -185,7 +_,7 @@
             boolean flag = false;
 
             for (Direction direction : Direction.values()) {
-                if (p_49270_.getBlockState(blockpos$mutableblockpos.set(p_49271_).move(direction)).is(Blocks.OBSIDIAN)) {
+                if (p_49270_.getBlockState(blockpos$mutableblockpos.set(p_49271_).move(direction)).isPortalFrame(p_49270_, blockpos$mutableblockpos)) {
                     flag = true;
                     break;
                 }
@@ -201,4 +_,12 @@
             }
         }
     }
+
+    // CraftBukkit start
+    protected void fireExtinguished(net.minecraft.world.level.LevelAccessor world, BlockPos position) {
+        if (!CraftEventFactory.callBlockFadeEvent(world, position, Blocks.AIR.defaultBlockState()).isCancelled()) {
+            world.removeBlock(position, false);
+        }
+    }
+    // CraftBukkit end
 }
