--- a/net/minecraft/world/level/block/BeaconBlock.java
+++ b/net/minecraft/world/level/block/BeaconBlock.java
@@ -49,8 +_,7 @@
         if (p_49433_.isClientSide) {
             return InteractionResult.SUCCESS;
         } else {
-            if (p_49433_.getBlockEntity(p_49434_) instanceof BeaconBlockEntity beaconblockentity) {
-                p_49435_.openMenu(beaconblockentity);
+            if (p_49433_.getBlockEntity(p_49434_) instanceof BeaconBlockEntity beaconblockentity && p_49435_.openMenu(beaconblockentity).isPresent()) { // Paper - Fix InventoryOpenEvent cancellation
                 p_49435_.awardStat(Stats.INTERACT_WITH_BEACON);
             }
 
