--- a/net/minecraft/world/level/block/BedBlock.java
+++ b/net/minecraft/world/level/block/BedBlock.java
@@ -26,6 +_,7 @@
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.CollisionGetter;
+import net.minecraft.world.level.ExplosionDamageCalculator;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.ScheduledTickAccess;
@@ -79,41 +_,51 @@
     protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hitResult) {
         if (level.isClientSide()) {
             return InteractionResult.SUCCESS_SERVER;
-        }
-
-        if (state.getValue(PART) != BedPart.HEAD) {
-            pos = pos.relative(state.getValue(FACING));
-            state = level.getBlockState(pos);
-            if (!state.is(this)) {
-                return InteractionResult.CONSUME;
-            }
-        }
-
-        BedRule bedRule = level.environmentAttributes().getValue(EnvironmentAttributes.BED_RULE, pos);
-        if (bedRule.explodes()) {
-            bedRule.errorMessage().ifPresent(player::sendOverlayMessage);
-            level.removeBlock(pos, false);
-            BlockPos blockPos = pos.relative(state.getValue(FACING).getOpposite());
-            if (level.getBlockState(blockPos).is(this)) {
-                level.removeBlock(blockPos, false);
-            }
-
-            Vec3 boomPos = Vec3.atCenterOf(pos);
-            level.explode(null, level.damageSources().badRespawnPointExplosion(boomPos), null, boomPos, 5.0F, true, Level.ExplosionInteraction.BLOCK);
-            return InteractionResult.SUCCESS_SERVER;
-        } else if (state.getValue(OCCUPIED)) {
-            if (!this.kickVillagerOutOfBed(level, pos)) {
-                player.sendOverlayMessage(Component.translatable("block.minecraft.bed.occupied"));
-            }
-
-            return InteractionResult.SUCCESS_SERVER;
         } else {
-            player.startSleepInBed(pos).ifLeft(problem -> {
-                if (problem.message() != null) {
-                    player.sendOverlayMessage(problem.message());
-                }
-            });
-            return InteractionResult.SUCCESS_SERVER;
+            if (state.getValue(PART) != BedPart.HEAD) {
+                pos = pos.relative(state.getValue(FACING));
+                state = level.getBlockState(pos);
+                if (!state.is(this)) {
+                    return InteractionResult.CONSUME;
+                }
+            }
+
+            BedRule bedRule = level.environmentAttributes().getValue(EnvironmentAttributes.BED_RULE, pos);
+            // CraftBukkit - moved world and biome check into EntityHuman
+            if (false && bedRule.explodes()) {
+                bedRule.errorMessage().ifPresent(player::sendOverlayMessage);
+                level.removeBlock(pos, false);
+                BlockPos blockPos = pos.relative(state.getValue(FACING).getOpposite());
+                if (level.getBlockState(blockPos).is(this)) {
+                    level.removeBlock(blockPos, false);
+                }
+
+                Vec3 boomPos = pos.getCenter();
+                level.explode(null, level.damageSources().badRespawnPointExplosion(boomPos), null, boomPos, 5.0F, true, Level.ExplosionInteraction.BLOCK);
+                return InteractionResult.SUCCESS_SERVER;
+            } else if (state.getValue(OCCUPIED)) {
+                if (!this.kickVillagerOutOfBed(level, pos)) {
+                    player.sendOverlayMessage(Component.translatable("block.minecraft.bed.occupied"));
+                }
+
+                return InteractionResult.SUCCESS_SERVER;
+            } else {
+                // CraftBukkit start
+                BlockState finalBlockState = state;
+                BlockPos finalPos = pos;
+                // CraftBukkit end
+                player.startSleepInBed(pos).ifLeft(problem -> {
+                    // CraftBukkit start - handling bed explosion from below here
+                    if (bedRule.explodes()) {
+                        BedBlock.this.explodeBed(finalBlockState, level, finalPos);
+                    } else
+                    // CraftBukkit end
+                    if (problem.message() != null) {
+                        player.sendOverlayMessage(problem.message());
+                    }
+                });
+                return InteractionResult.SUCCESS_SERVER;
+            }
         }
     }
 
@@ -126,6 +_,27 @@
         villagers.get(0).stopSleeping();
         return true;
     }
+
+    // CraftBukkit start
+    private InteractionResult explodeBed(BlockState state, Level level, BlockPos pos) {
+        {
+            {
+                org.bukkit.block.BlockState blockState = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos).getState(); // CraftBukkit - capture BlockState before remove block
+                level.removeBlock(pos, false);
+                BlockPos blockpos1 = pos.relative(((Direction) state.getValue(BedBlock.FACING)).getOpposite());
+
+                if (level.getBlockState(blockpos1).getBlock() == this) {
+                    level.removeBlock(blockpos1, false);
+                }
+
+                Vec3 vec3 = Vec3.atCenterOf(pos);
+
+                level.explode((Entity) null, level.damageSources().badRespawnPointExplosion(vec3, blockState), (ExplosionDamageCalculator) null, vec3, 5.0F, true, Level.ExplosionInteraction.BLOCK); // CraftBukkit - add state
+                return InteractionResult.SUCCESS;
+            }
+        }
+    }
+    // CraftBukkit end
 
     @Override
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, double fallDistance) {
