--- a/net/minecraft/world/level/block/BedBlock.java
+++ b/net/minecraft/world/level/block/BedBlock.java
@@ -2,7 +_,6 @@
 
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.List;
 import java.util.Optional;
 import javax.annotation.Nullable;
@@ -41,6 +_,7 @@
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.apache.commons.lang3.ArrayUtils;
+import org.bukkit.craftbukkit.block.CraftBlock;
 
 public class BedBlock extends HorizontalDirectionalBlock implements EntityBlock {
     public static final MapCodec<BedBlock> CODEC = RecordCodecBuilder.mapCodec(
@@ -108,16 +_,60 @@
 
                 return InteractionResult.SUCCESS;
             } else {
+                // CraftBukkit start
+                BlockState finaliblockdata = p_49515_;
+                BlockPos finalblockposition = p_49517_;
+                // CraftBukkit end
                 p_49518_.startSleepInBed(p_49517_).ifLeft(p_49477_ -> {
-                    if (p_49477_.getMessage() != null) {
-                        p_49518_.displayClientMessage(p_49477_.getMessage(), true);
-                    }
+                    // Paper start - PlayerBedFailEnterEvent
+                    if (p_49477_ != null) {
+                        io.papermc.paper.event.player.PlayerBedFailEnterEvent event = new io.papermc.paper.event.player.PlayerBedFailEnterEvent(
+                                (org.bukkit.entity.Player) p_49518_.getBukkitEntity(),
+                                io.papermc.paper.event.player.PlayerBedFailEnterEvent.FailReason.values()[p_49477_.ordinal()],
+                                org.bukkit.craftbukkit.block.CraftBlock.at(p_49516_, finalblockposition),
+                                !p_49516_.dimensionType().bedWorks(),
+                                io.papermc.paper.adventure.PaperAdventure.asAdventure(p_49477_.getMessage()));
+                        if (!event.callEvent()) {
+                            return;
+                        }
+                        // Paper end - PlayerBedFailEnterEvent
+                        // CraftBukkit start - handling bed explosion from below here
+                        if (event.getWillExplode()) { // Paper - PlayerBedFailEnterEvent
+                            this.explodeBed(finaliblockdata, p_49516_, finalblockposition);
+                        } else
+                            // CraftBukkit end
+                            if (p_49477_.getMessage() != null) {
+                                final net.kyori.adventure.text.Component message = event.getMessage(); // Paper - PlayerBedFailEnterEvent
+                                if (message != null) p_49518_.displayClientMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(message), true); // Paper - PlayerBedFailEnterEvent
+                            }
+                    } // Paper - PlayerBedFailEnterEvent
                 });
                 return InteractionResult.SUCCESS;
             }
         }
     }
 
+    // CraftBukkit start
+    private InteractionResult explodeBed(BlockState blockState, Level world, BlockPos blockPos) {
+        {
+            {
+                org.bukkit.block.BlockState blockState2 = CraftBlock.at(world, blockPos).getState(); // CraftBukkit - capture BlockState before remove block
+                world.removeBlock(blockPos, false);
+                BlockPos blockPos1 = blockPos.relative(blockState.getValue(FACING).getOpposite());
+
+                if (world.getBlockState(blockPos1).getBlock() == this) {
+                    world.removeBlock(blockPos1, false);
+                }
+
+                Vec3 vec3 = blockPos.getCenter();
+
+                world.explode(null, world.damageSources().badRespawnPointExplosionCB(vec3, blockState2), null, vec3, 5.0F, true, Level.ExplosionInteraction.BLOCK); // CraftBukkit - add state
+                return InteractionResult.SUCCESS;
+            }
+        }
+    }
+    // CraftBukkit end
+
     public static boolean canSetSpawn(Level p_49489_) {
         return p_49489_.dimensionType().bedWorks();
     }
@@ -134,7 +_,7 @@
 
     @Override
     public void fallOn(Level p_152169_, BlockState p_152170_, BlockPos p_152171_, Entity p_152172_, float p_152173_) {
-        super.fallOn(p_152169_, p_152170_, p_152171_, p_152172_, p_152173_ * 0.5F);
+        super.fallOn(p_152169_, p_152170_, p_152171_, p_152172_, p_152173_); // Purpur
     }
 
     @Override
@@ -308,6 +_,11 @@
         if (!p_49499_.isClientSide) {
             BlockPos blockpos = p_49500_.relative(p_49501_.getValue(FACING));
             p_49499_.setBlock(blockpos, p_49501_.setValue(PART, BedPart.HEAD), 3);
+            // CraftBukkit start - SPIGOT-7315: Don't updated if we capture block states
+            if (p_49499_.captureBlockStates) {
+                return;
+            }
+            // CraftBukkit end
             p_49499_.blockUpdated(p_49500_, Blocks.AIR);
             p_49501_.updateNeighbourShapes(p_49499_, p_49500_, 3);
         }
