--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -108,7 +_,7 @@
             for (Bee bee : list) {
                 if (bee.getTarget() == null) {
                     Player player = Util.getRandom(list1, p_49650_.random);
-                    bee.setTarget(player);
+                    bee.setTarget(player, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER, true); // CraftBukkit
                 }
             }
         }
@@ -126,7 +_,7 @@
         boolean flag = false;
         if (i >= 5) {
             Item item = p_316844_.getItem();
-            if (p_316844_.is(Items.SHEARS)) {
+            if (p_316844_.canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SHEARS_HARVEST)) {
                 p_316306_.playSound(p_316824_, p_316824_.getX(), p_316824_.getY(), p_316824_.getZ(), SoundEvents.BEEHIVE_SHEAR, SoundSource.BLOCKS, 1.0F, 1.0F);
                 dropHoneycomb(p_316306_, p_316497_);
                 p_316844_.hurtAndBreak(1, p_316824_, LivingEntity.getSlotForHand(p_316436_));
