--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -2,13 +_,11 @@
 
 import java.util.List;
 import java.util.function.BiConsumer;
-import net.minecraft.advancements.triggers.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
@@ -81,8 +_,8 @@
     }
 
     @Override
-    public void playerDestroy(Level level, Player player, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, ItemStack destroyedWith) {
-        super.playerDestroy(level, player, pos, state, blockEntity, destroyedWith);
+    public void playerDestroy(Level level, Player player, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, ItemStack destroyedWith,  boolean includeDrops, boolean dropExp) {
+        super.playerDestroy(level, player, pos, state, blockEntity, destroyedWith, includeDrops, dropExp); // Paper - fix drops not preventing stats/food exhaustion
         if (!level.isClientSide() && blockEntity instanceof BeehiveBlockEntity beehiveBlockEntity) {
             if (!EnchantmentHelper.hasTag(destroyedWith, EnchantmentTags.PREVENTS_BEE_SPAWNS_WHEN_MINING)) {
                 beehiveBlockEntity.emptyAllLivingFromHive(player, state, BeehiveBlockEntity.BeeReleaseStatus.EMERGENCY);
@@ -90,7 +_,7 @@
                 this.angerNearbyBees(level, pos);
             }
 
-            CriteriaTriggers.BEE_NEST_DESTROYED.trigger((ServerPlayer)player, state, destroyedWith, beehiveBlockEntity.getOccupantCount());
+            // CriteriaTriggers.BEE_NEST_DESTROYED.trigger((ServerPlayer)player, state, destroyedWith, beehiveBlockEntity.getOccupantCount()); // Paper - Trigger bee_nest_destroyed trigger in the correct place; moved until after items are dropped
         }
     }
 
@@ -112,6 +_,7 @@
             for (Bee bee : beesToAnger) {
                 if (bee.getTarget() == null) {
                     Player angerTarget = Util.getRandom(playersToBeAngryAt, level.getRandom());
+                    bee.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER); // CraftBukkit
                     bee.setTarget(angerTarget);
                 }
             }
@@ -134,8 +_,20 @@
         boolean hiveEmptied = false;
         if (honeyLevel >= 5) {
             Item item = itemStack.getItem();
-            if (level instanceof ServerLevel serverLevel && itemStack.is(Items.SHEARS)) {
-                dropHoneycomb(serverLevel, itemStack, state, level.getBlockEntity(pos), player, pos);
+            if (level instanceof ServerLevel serverLevel && itemStack.canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SHEARS_HARVEST)) {
+                // Paper start - Add PlayerShearBlockEvent
+                List<org.bukkit.inventory.ItemStack> drops = new java.util.ArrayList<>();
+                dropFromBlockInteractLootTable(
+                        serverLevel, BuiltInLootTables.HARVEST_BEEHIVE, state, level.getBlockEntity(pos), itemStack, player, (_, stack) -> drops.add(org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack))
+                );
+                io.papermc.paper.event.block.PlayerShearBlockEvent event = new io.papermc.paper.event.block.PlayerShearBlockEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack), org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand), drops);
+                if (!event.callEvent()) {
+                    return InteractionResult.PASS;
+                }
+                for (org.bukkit.inventory.ItemStack stack : event.getDrops()) {
+                    popResource(serverLevel, pos, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(stack));
+                }
+                // Paper end - Add PlayerShearBlockEvent
                 level.playSound(null, player.getX(), player.getY(), player.getZ(), SoundEvents.BEEHIVE_SHEAR, SoundSource.BLOCKS, 1.0F, 1.0F);
                 itemStack.hurtAndBreak(1, player, hand.asEquipmentSlot());
                 hiveEmptied = true;
@@ -189,7 +_,7 @@
     }
 
     public void resetHoneyLevel(Level level, BlockState state, BlockPos pos) {
-        level.setBlockAndUpdate(pos, state.setValue(HONEY_LEVEL, 0));
+        level.setBlock(pos, state.setValue(HONEY_LEVEL, 0), 3);
     }
 
     @Override
