--- a/net/minecraft/world/level/block/BellBlock.java
+++ b/net/minecraft/world/level/block/BellBlock.java
@@ -139,6 +_,11 @@
             if (p_152192_ == null) {
                 p_152192_ = p_152190_.getBlockState(p_152191_).getValue(FACING);
             }
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBellRingEvent(p_152190_, p_152191_, p_152192_, p_152189_)) {
+                return false;
+            }
+            // CraftBukkit end
 
             ((BellBlockEntity)blockentity).onHit(p_152192_);
             p_152190_.playSound(null, p_152191_, SoundEvents.BELL_BLOCK, SoundSource.BLOCKS, 2.0F, 1.0F);
