--- a/net/minecraft/world/level/block/BellBlock.java
+++ b/net/minecraft/world/level/block/BellBlock.java
@@ -65,11 +_,16 @@
     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block block, @Nullable Orientation orientation, boolean movedByPiston) {
         boolean signal = level.hasNeighborSignal(pos);
         if (signal != state.getValue(POWERED)) {
+            // Paper start - Call BlockRedstoneEvent
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBinaryRedstoneChange(level, pos, signal)) {
+                return;
+            }
+            // Paper end - Call BlockRedstoneEvent
             if (signal) {
                 this.attemptToRing(level, pos, null);
             }
 
-            level.setBlockAndUpdate(pos, state.setValue(POWERED, signal));
+            level.setBlock(pos, state.setValue(POWERED, signal), 3);
         }
     }
 
@@ -126,6 +_,11 @@
                 direction = level.getBlockState(pos).getValue(FACING);
             }
 
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBellRingEvent(level, pos, direction, ringingEntity)) {
+                return false;
+            }
+            // CraftBukkit end
             bellBlockEntity.onHit(direction);
             level.playSound(null, pos, SoundEvents.BELL_BLOCK, SoundSource.BLOCKS, 2.0F, 1.0F);
             level.gameEvent(ringingEntity, GameEvent.BLOCK_CHANGE, pos);
