--- a/net/minecraft/world/level/block/BlastFurnaceBlock.java
+++ b/net/minecraft/world/level/block/BlastFurnaceBlock.java
@@ -45,8 +_,7 @@
     @Override
     protected void openContainer(Level p_49777_, BlockPos p_49778_, Player p_49779_) {
         BlockEntity blockentity = p_49777_.getBlockEntity(p_49778_);
-        if (blockentity instanceof BlastFurnaceBlockEntity) {
-            p_49779_.openMenu((MenuProvider)blockentity);
+        if (blockentity instanceof BlastFurnaceBlockEntity && p_49779_.openMenu((MenuProvider)blockentity).isPresent()) { // Paper - Fix InventoryOpenEvent cancellation
             p_49779_.awardStat(Stats.INTERACT_WITH_BLAST_FURNACE);
         }
     }
