--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -46,22 +_,12 @@
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.level.material.MapColor;
 import net.minecraft.world.level.material.PushReaction;
-import net.minecraft.world.phys.AABB;
 
 public class Blocks {
     private static final BlockBehaviour.StatePredicate NOT_CLOSED_SHULKER = (state, level, pos) -> level.getBlockEntity(pos) instanceof ShulkerBoxBlockEntity shulkerBoxBlockEntity
         ? shulkerBoxBlockEntity.isClosed()
         : true;
     private static final BlockBehaviour.StatePredicate NOT_EXTENDED_PISTON = (state, level, pos) -> !state.getValue(PistonBaseBlock.EXTENDED);
-    private static final BlockBehaviour.StateArgumentPredicate<AABB> NEAR_PLANE_INTERSECTS_OUTLINE = (state, level, blockPos, nearPlaneBox) -> {
-        for (AABB outlineBox : state.getOcclusionShape().toAabbs()) {
-            if (outlineBox.move(blockPos).intersects(nearPlaneBox)) {
-                return true;
-            }
-        }
-
-        return false;
-    };
     public static final Block AIR = register(BlockItemIds.AIR, AirBlock::new, BlockBehaviour.Properties.of().replaceable().noCollision().noLootTable().air());
     public static final Block STONE = register(
         BlockItemIds.STONE,
@@ -212,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SPRUCE_SAPLING = register(
         BlockItemIds.SPRUCE_SAPLING,
@@ -223,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BIRCH_SAPLING = register(
         BlockItemIds.BIRCH_SAPLING,
@@ -234,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block JUNGLE_SAPLING = register(
         BlockItemIds.JUNGLE_SAPLING,
@@ -245,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block ACACIA_SAPLING = register(
         BlockItemIds.ACACIA_SAPLING,
@@ -256,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CHERRY_SAPLING = register(
         BlockItemIds.CHERRY_SAPLING,
@@ -267,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.CHERRY_SAPLING)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block DARK_OAK_SAPLING = register(
         BlockItemIds.DARK_OAK_SAPLING,
@@ -278,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block PALE_OAK_SAPLING = register(
         BlockItemIds.PALE_OAK_SAPLING,
@@ -289,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block MANGROVE_PROPAGULE = register(
         BlockItemIds.MANGROVE_PROPAGULE,
@@ -301,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block POPLAR_SAPLING = register(
         BlockItemIds.POPLAR_SAPLING,
@@ -312,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BEDROCK = register(
         BlockItemIds.BEDROCK,
@@ -331,7 +_,7 @@
             .replaceable()
             .noCollision()
             .strength(100.0F)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .noLootTable()
             .liquid()
             .sound(SoundType.EMPTY)
@@ -346,7 +_,7 @@
             .randomTicks()
             .strength(100.0F)
             .lightLevel(state -> 15)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .noLootTable()
             .liquid()
             .sound(SoundType.EMPTY)
@@ -364,7 +_,7 @@
             .instrument(NoteBlockInstrument.SNARE)
             .strength(0.25F)
             .sound(SoundType.SUSPICIOUS_SAND)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block RED_SAND = register(
         BlockItemIds.RED_SAND,
@@ -384,7 +_,7 @@
             .instrument(NoteBlockInstrument.SNARE)
             .strength(0.25F)
             .sound(SoundType.SUSPICIOUS_GRAVEL)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block GOLD_ORE = register(
         BlockItemIds.GOLD_ORE,
@@ -464,6 +_,7 @@
             .sound(SoundType.MANGROVE_ROOTS)
             .noOcclusion()
             .isSuffocating(Blocks::never)
+            .isViewBlocking(Blocks::never)
             .noOcclusion()
             .ignitedByLava()
     );
@@ -654,8 +_,9 @@
             .noOcclusion()
             .isValidSpawn(Blocks::ocelotOrParrot)
             .isSuffocating(Blocks::never)
+            .isViewBlocking(Blocks::never)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .isRedstoneConductor(Blocks::never)
     );
     public static final Block DARK_OAK_LEAVES = register(
@@ -672,8 +_,9 @@
             .noOcclusion()
             .isValidSpawn(Blocks::ocelotOrParrot)
             .isSuffocating(Blocks::never)
+            .isViewBlocking(Blocks::never)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .isRedstoneConductor(Blocks::never)
     );
     public static final Block MANGROVE_LEAVES = register(
@@ -740,6 +_,7 @@
             .isValidSpawn(Blocks::never)
             .isRedstoneConductor(Blocks::never)
             .isSuffocating(Blocks::never)
+            .isViewBlocking(Blocks::never)
     );
     public static final Block LAPIS_ORE = register(
         BlockItemIds.LAPIS_ORE,
@@ -788,7 +_,7 @@
             .fallDistanceReduction(0.5F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block STRAW_BED = register(
         BlockItemIds.STRAW_BED,
@@ -799,7 +_,7 @@
             .strength(0.2F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block POWERED_RAIL = register(
         BlockItemIds.POWERED_RAIL, PoweredRailBlock::new, BlockBehaviour.Properties.of().noCollision().strength(0.7F).sound(SoundType.METAL)
@@ -818,7 +_,7 @@
             .noCollision()
             .requiresCorrectToolForDrops()
             .strength(4.0F)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SHORT_GRASS = register(
         BlockItemIds.SHORT_GRASS,
@@ -831,7 +_,7 @@
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XYZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block FERN = register(
         BlockItemIds.FERN,
@@ -844,7 +_,7 @@
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XYZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block DEAD_BUSH = register(
         BlockItemIds.DEAD_BUSH,
@@ -856,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BUSH = register(
         BlockItemIds.BUSH,
@@ -868,19 +_,19 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block RED_SHRUB = register(
         BlockItemIds.RED_SHRUB,
         BushBlock::new,
         BlockBehaviour.Properties.of()
-            .mapColor(MapColor.CRIMSON_NYLIUM)
+            .mapColor(MapColor.PLANT)
             .replaceable()
             .noCollision()
             .instabreak()
             .sound(SoundType.RED_SHRUB)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SHORT_DRY_GRASS = register(
         BlockItemIds.SHORT_DRY_GRASS,
@@ -893,7 +_,7 @@
             .sound(SoundType.GRASS)
             .ignitedByLava()
             .offsetType(BlockBehaviour.OffsetType.XYZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block TALL_DRY_GRASS = register(
         BlockItemIds.TALL_DRY_GRASS,
@@ -906,7 +_,7 @@
             .sound(SoundType.GRASS)
             .ignitedByLava()
             .offsetType(BlockBehaviour.OffsetType.XYZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SEAGRASS = register(
         BlockItemIds.SEAGRASS,
@@ -917,7 +_,7 @@
             .noCollision()
             .instabreak()
             .sound(SoundType.WET_GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block TALL_SEAGRASS = register(
         BlockIds.TALL_SEAGRASS,
@@ -929,13 +_,13 @@
             .instabreak()
             .sound(SoundType.WET_GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block PISTON = register(BlockItemIds.PISTON, p -> new PistonBaseBlock(false, p), pistonProperties());
     public static final Block PISTON_HEAD = register(
         BlockIds.PISTON_HEAD,
         PistonHeadBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.STONE).strength(1.5F).noLootTable().pushReaction(PushReaction.IMMOVEABLE)
+        BlockBehaviour.Properties.of().mapColor(MapColor.STONE).strength(1.5F).noLootTable().pushReaction(PushReaction.BLOCK)
     );
     public static final ColorCollection<Block> WOOL = ColorCollection.registerBlocks(
         BlockItemIds.WOOL,
@@ -967,7 +_,8 @@
             .isValidSpawn(Blocks::never)
             .isRedstoneConductor(Blocks::never)
             .isSuffocating(Blocks::never)
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .isViewBlocking(Blocks::never)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block DANDELION = register(
         BlockItemIds.DANDELION,
@@ -978,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block GOLDEN_DANDELION = register(
         BlockItemIds.GOLDEN_DANDELION,
@@ -989,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block TORCHFLOWER = register(
         BlockItemIds.TORCHFLOWER,
@@ -1000,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block POPPY = register(
         BlockItemIds.POPPY,
@@ -1011,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BLUE_ORCHID = register(
         BlockItemIds.BLUE_ORCHID,
@@ -1022,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block ALLIUM = register(
         BlockItemIds.ALLIUM,
@@ -1033,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block AZURE_BLUET = register(
         BlockItemIds.AZURE_BLUET,
@@ -1044,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block RED_TULIP = register(
         BlockItemIds.RED_TULIP,
@@ -1055,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block ORANGE_TULIP = register(
         BlockItemIds.ORANGE_TULIP,
@@ -1066,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block WHITE_TULIP = register(
         BlockItemIds.WHITE_TULIP,
@@ -1077,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block PINK_TULIP = register(
         BlockItemIds.PINK_TULIP,
@@ -1088,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block OXEYE_DAISY = register(
         BlockItemIds.OXEYE_DAISY,
@@ -1099,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CORNFLOWER = register(
         BlockItemIds.CORNFLOWER,
@@ -1110,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block WITHER_ROSE = register(
         BlockItemIds.WITHER_ROSE,
@@ -1121,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block LILY_OF_THE_VALLEY = register(
         BlockItemIds.LILY_OF_THE_VALLEY,
@@ -1132,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BROWN_MUSHROOM = register(
         BlockItemIds.BROWN_MUSHROOM,
@@ -1145,7 +_,7 @@
             .sound(SoundType.GRASS)
             .lightLevel(state -> 1)
             .postProcess(Blocks::postProcessSelf)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block RED_MUSHROOM = register(
         BlockItemIds.RED_MUSHROOM,
@@ -1157,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .postProcess(Blocks::postProcessSelf)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block GOLD_BLOCK = register(
         BlockItemIds.GOLD_BLOCK,
@@ -1341,17 +_,16 @@
             .instrument(NoteBlockInstrument.BASEDRUM)
             .requiresCorrectToolForDrops()
             .strength(50.0F, 1200.0F)
-            .pushReaction(PushReaction.IMMOVEABLE)
     );
     public static final Block TORCH = register(
         BlockItemIds.TORCH,
         p -> new TorchBlock(ParticleTypes.FLAME, p),
-        BlockBehaviour.Properties.of().noCollision().instabreak().lightLevel(state -> 14).sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().noCollision().instabreak().lightLevel(state -> 14).sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block WALL_TORCH = register(
         BlockIds.WALL_TORCH,
         p -> new WallTorchBlock(ParticleTypes.FLAME, p),
-        wallVariant(TORCH, true).noCollision().instabreak().lightLevel(state -> 14).sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        wallVariant(TORCH, true).noCollision().instabreak().lightLevel(state -> 14).sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block FIRE = register(
         BlockIds.FIRE,
@@ -1363,7 +_,7 @@
             .instabreak()
             .lightLevel(state -> 15)
             .sound(SoundType.WOOL)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SOUL_FIRE = register(
         BlockIds.SOUL_FIRE,
@@ -1375,7 +_,7 @@
             .instabreak()
             .lightLevel(state -> 10)
             .sound(SoundType.WOOL)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SPAWNER = register(
         BlockItemIds.SPAWNER,
@@ -1400,7 +_,7 @@
         BlockBehaviour.Properties.of().mapColor(MapColor.WOOD).instrument(NoteBlockInstrument.BASS).strength(2.5F).sound(SoundType.WOOD).ignitedByLava()
     );
     public static final Block REDSTONE_WIRE = register(
-        BlockItemIds.REDSTONE_DUST, RedstoneWireBlock::new, BlockBehaviour.Properties.of().noCollision().instabreak().pushReaction(PushReaction.POPPED)
+        BlockItemIds.REDSTONE_DUST, RedStoneWireBlock::new, BlockBehaviour.Properties.of().noCollision().instabreak().pushReaction(PushReaction.DESTROY)
     );
     public static final Block DIAMOND_ORE = register(
         BlockItemIds.DIAMOND_ORE,
@@ -1430,12 +_,18 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.CROP)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block FARMLAND = register(
         BlockItemIds.FARMLAND,
         p -> new FarmlandBlock(DIRT, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.DIRT).randomTicks().strength(0.6F).sound(SoundType.GRAVEL).isSuffocating(Blocks::always)
+        BlockBehaviour.Properties.of()
+            .mapColor(MapColor.DIRT)
+            .randomTicks()
+            .strength(0.6F)
+            .sound(SoundType.GRAVEL)
+            .isViewBlocking(Blocks::always)
+            .isSuffocating(Blocks::always)
     );
     public static final Block FURNACE = register(
         BlockItemIds.FURNACE,
@@ -1565,12 +_,12 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block LADDER = register(
         BlockItemIds.LADDER,
         LadderBlock::new,
-        BlockBehaviour.Properties.of().forceSolidOff().strength(0.4F).sound(SoundType.LADDER).noOcclusion().pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().forceSolidOff().strength(0.4F).sound(SoundType.LADDER).noOcclusion().pushReaction(PushReaction.DESTROY)
     );
     public static final Block RAIL = register(
         BlockItemIds.RAIL, RailBlock::new, BlockBehaviour.Properties.of().noCollision().strength(0.7F).sound(SoundType.METAL)
@@ -1949,7 +_,7 @@
     public static final Block LEVER = register(
         BlockItemIds.LEVER,
         LeverBlock::new,
-        BlockBehaviour.Properties.of().noCollision().strength(0.5F).sound(SoundType.STONE).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().noCollision().strength(0.5F).sound(SoundType.STONE).pushReaction(PushReaction.DESTROY)
     );
     public static final Block STONE_PRESSURE_PLATE = register(
         BlockItemIds.STONE_PRESSURE_PLATE,
@@ -1960,12 +_,12 @@
             .instrument(NoteBlockInstrument.BASEDRUM)
             .noCollision()
             .strength(0.5F)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block IRON_DOOR = register(
         BlockItemIds.IRON_DOOR,
         p -> new DoorBlock(BlockSetType.IRON, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.METAL).strength(5.0F).noOcclusion().pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.METAL).strength(5.0F).noOcclusion().pushReaction(PushReaction.DESTROY)
     );
     public static final Block OAK_PRESSURE_PLATE = register(
         BlockItemIds.OAK_PRESSURE_PLATE,
@@ -1977,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SPRUCE_PRESSURE_PLATE = register(
         BlockItemIds.SPRUCE_PRESSURE_PLATE,
@@ -1989,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BIRCH_PRESSURE_PLATE = register(
         BlockItemIds.BIRCH_PRESSURE_PLATE,
@@ -2001,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block JUNGLE_PRESSURE_PLATE = register(
         BlockItemIds.JUNGLE_PRESSURE_PLATE,
@@ -2013,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block ACACIA_PRESSURE_PLATE = register(
         BlockItemIds.ACACIA_PRESSURE_PLATE,
@@ -2025,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CHERRY_PRESSURE_PLATE = register(
         BlockItemIds.CHERRY_PRESSURE_PLATE,
@@ -2037,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block DARK_OAK_PRESSURE_PLATE = register(
         BlockItemIds.DARK_OAK_PRESSURE_PLATE,
@@ -2049,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block PALE_OAK_PRESSURE_PLATE = register(
         BlockItemIds.PALE_OAK_PRESSURE_PLATE,
@@ -2061,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block MANGROVE_PRESSURE_PLATE = register(
         BlockItemIds.MANGROVE_PRESSURE_PLATE,
@@ -2073,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block POPLAR_PRESSURE_PLATE = register(
         BlockItemIds.POPLAR_PRESSURE_PLATE,
@@ -2085,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BAMBOO_PRESSURE_PLATE = register(
         BlockItemIds.BAMBOO_PRESSURE_PLATE,
@@ -2097,7 +_,7 @@
             .noCollision()
             .strength(0.5F)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block REDSTONE_ORE = register(
         BlockItemIds.REDSTONE_ORE,
@@ -2118,12 +_,12 @@
     public static final Block REDSTONE_TORCH = register(
         BlockItemIds.REDSTONE_TORCH,
         RedstoneTorchBlock::new,
-        BlockBehaviour.Properties.of().noCollision().instabreak().lightLevel(litBlockEmission(7)).sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().noCollision().instabreak().lightLevel(litBlockEmission(7)).sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block REDSTONE_WALL_TORCH = register(
         BlockIds.REDSTONE_WALL_TORCH,
         RedstoneWallTorchBlock::new,
-        wallVariant(REDSTONE_TORCH, true).noCollision().instabreak().lightLevel(litBlockEmission(7)).sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        wallVariant(REDSTONE_TORCH, true).noCollision().instabreak().lightLevel(litBlockEmission(7)).sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block STONE_BUTTON = register(BlockItemIds.STONE_BUTTON, p -> new ButtonBlock(BlockSetType.STONE, 20, p), buttonProperties());
     public static final Block SNOW = register(
@@ -2137,8 +_,8 @@
             .strength(0.1F)
             .requiresCorrectToolForDrops()
             .sound(SoundType.SNOW)
-            .isViewBlocking((state, level, pos, aabb) -> state.getValue(SnowLayerBlock.LAYERS) >= 8)
-            .pushReaction(PushReaction.POPPED)
+            .isViewBlocking((state, level, pos) -> state.getValue(SnowLayerBlock.LAYERS) >= 8)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block ICE = register(
         BlockItemIds.ICE,
@@ -2159,7 +_,7 @@
     public static final Block CACTUS = register(
         BlockItemIds.CACTUS,
         CactusBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).randomTicks().strength(0.4F).sound(SoundType.WOOL).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).randomTicks().strength(0.4F).sound(SoundType.WOOL).pushReaction(PushReaction.DESTROY)
     );
     public static final Block CACTUS_FLOWER = register(
         BlockItemIds.CACTUS_FLOWER,
@@ -2170,7 +_,7 @@
             .instabreak()
             .ignitedByLava()
             .sound(SoundType.CACTUS_FLOWER)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CLAY = register(
         BlockItemIds.CLAY, BlockBehaviour.Properties.of().mapColor(MapColor.CLAY).instrument(NoteBlockInstrument.FLUTE).strength(0.6F).sound(SoundType.GRAVEL)
@@ -2184,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block JUKEBOX = register(
         BlockItemIds.JUKEBOX,
@@ -2223,6 +_,7 @@
             .sound(SoundType.SOUL_SAND)
             .isValidSpawn(Blocks::always)
             .isRedstoneConductor(Blocks::always)
+            .isViewBlocking(Blocks::always)
             .isSuffocating(Blocks::always)
             .postProcess(Blocks::postProcessAbove)
     );
@@ -2252,22 +_,22 @@
     public static final Block SOUL_TORCH = register(
         BlockItemIds.SOUL_TORCH,
         p -> new TorchBlock(ParticleTypes.SOUL_FIRE_FLAME, p),
-        BlockBehaviour.Properties.of().noCollision().instabreak().lightLevel(state -> 10).sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().noCollision().instabreak().lightLevel(state -> 10).sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block SOUL_WALL_TORCH = register(
         BlockIds.SOUL_WALL_TORCH,
         p -> new WallTorchBlock(ParticleTypes.SOUL_FIRE_FLAME, p),
-        wallVariant(SOUL_TORCH, true).noCollision().instabreak().lightLevel(state -> 10).sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        wallVariant(SOUL_TORCH, true).noCollision().instabreak().lightLevel(state -> 10).sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block COPPER_TORCH = register(
         BlockItemIds.COPPER_TORCH,
         p -> new TorchBlock(ParticleTypes.COPPER_FIRE_FLAME, p),
-        BlockBehaviour.Properties.of().noCollision().instabreak().lightLevel(state -> 14).sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().noCollision().instabreak().lightLevel(state -> 14).sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block COPPER_WALL_TORCH = register(
         BlockIds.COPPER_WALL_TORCH,
         p -> new WallTorchBlock(ParticleTypes.COPPER_FIRE_FLAME, p),
-        wallVariant(COPPER_TORCH, true).noCollision().instabreak().lightLevel(state -> 14).sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        wallVariant(COPPER_TORCH, true).noCollision().instabreak().lightLevel(state -> 14).sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block GLOWSTONE = register(
         BlockItemIds.GLOWSTONE,
@@ -2288,7 +_,7 @@
             .strength(-1.0F)
             .sound(SoundType.GLASS)
             .lightLevel(state -> 11)
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block CARVED_PUMPKIN = register(
         BlockItemIds.CARVED_PUMPKIN,
@@ -2298,7 +_,7 @@
             .strength(1.0F)
             .sound(SoundType.WOOD)
             .isValidSpawn(Blocks::always)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block JACK_O_LANTERN = register(
         BlockItemIds.JACK_O_LANTERN,
@@ -2309,13 +_,15 @@
             .sound(SoundType.WOOD)
             .lightLevel(state -> 15)
             .isValidSpawn(Blocks::always)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CAKE = register(
-        BlockItemIds.CAKE, CakeBlock::new, BlockBehaviour.Properties.of().forceSolidOn().strength(0.5F).sound(SoundType.WOOL).pushReaction(PushReaction.POPPED)
+        BlockItemIds.CAKE,
+        CakeBlock::new,
+        BlockBehaviour.Properties.of().forceSolidOn().strength(0.5F).sound(SoundType.WOOL).pushReaction(PushReaction.DESTROY)
     );
     public static final Block REPEATER = register(
-        BlockItemIds.REPEATER, RepeaterBlock::new, BlockBehaviour.Properties.of().instabreak().sound(SoundType.STONE).pushReaction(PushReaction.POPPED)
+        BlockItemIds.REPEATER, RepeaterBlock::new, BlockBehaviour.Properties.of().instabreak().sound(SoundType.STONE).pushReaction(PushReaction.DESTROY)
     );
     public static final ColorCollection<Block> STAINED_GLASS = ColorCollection.registerBlocks(
         BlockItemIds.STAINED_GLASS,
@@ -2330,6 +_,7 @@
             .isValidSpawn(Blocks::never)
             .isRedstoneConductor(Blocks::never)
             .isSuffocating(Blocks::never)
+            .isViewBlocking(Blocks::never)
     );
     public static final Block OAK_TRAPDOOR = register(
         BlockItemIds.OAK_TRAPDOOR,
@@ -2550,25 +_,25 @@
             .instrument(NoteBlockInstrument.DIDGERIDOO)
             .strength(1.0F)
             .sound(SoundType.WOOD)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block MELON = register(
         BlockItemIds.MELON,
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_LIGHT_GREEN).strength(1.0F).sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_LIGHT_GREEN).strength(1.0F).sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block ATTACHED_PUMPKIN_STEM = register(
         BlockIds.ATTACHED_PUMPKIN_STEM,
         p -> new AttachedStemBlock(
             BlockItemIds.PUMPKIN_CROP.block(), BlockItemIds.PUMPKIN.block(), BlockItemIds.PUMPKIN_CROP.item(), BlockTags.SUPPORTS_PUMPKIN_STEM, p
         ),
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().instabreak().sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().instabreak().sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block ATTACHED_MELON_STEM = register(
         BlockIds.ATTACHED_MELON_STEM,
         p -> new AttachedStemBlock(
             BlockItemIds.MELON_CROP.block(), BlockItemIds.MELON.block(), BlockItemIds.MELON_CROP.item(), BlockTags.SUPPORTS_MELON_STEM, p
         ),
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().instabreak().sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().instabreak().sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block PUMPKIN_STEM = register(
         BlockItemIds.PUMPKIN_CROP,
@@ -2586,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.HARD_CROP)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block MELON_STEM = register(
         BlockItemIds.MELON_CROP,
@@ -2604,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.HARD_CROP)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block VINE = register(
         BlockItemIds.VINE,
@@ -2617,7 +_,7 @@
             .strength(0.2F)
             .sound(SoundType.VINE)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block GLOW_LICHEN = register(
         BlockItemIds.GLOW_LICHEN,
@@ -2630,7 +_,7 @@
             .sound(SoundType.GLOW_LICHEN)
             .lightLevel(GlowLichenBlock.emission(7))
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block RESIN_CLUMP = register(
         BlockItemIds.RESIN_CLUMP,
@@ -2641,7 +_,7 @@
             .noCollision()
             .sound(SoundType.RESIN)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block OAK_FENCE_GATE = register(
         BlockItemIds.OAK_FENCE_GATE,
@@ -2664,7 +_,7 @@
     public static final Block LILY_PAD = register(
         BlockItemIds.LILY_PAD,
         LilyPadBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).instabreak().sound(SoundType.LILY_PAD).noOcclusion().pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).instabreak().sound(SoundType.LILY_PAD).noOcclusion().pushReaction(PushReaction.DESTROY)
     );
     public static final Block RESIN_BLOCK = register(
         BlockItemIds.RESIN_BLOCK,
@@ -2714,7 +_,7 @@
     public static final Block NETHER_WART = register(
         BlockItemIds.NETHER_WART,
         NetherWartBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_RED).noCollision().randomTicks().sound(SoundType.NETHER_WART).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_RED).noCollision().randomTicks().sound(SoundType.NETHER_WART).pushReaction(PushReaction.DESTROY)
     );
     public static final Block ENCHANTING_TABLE = register(
         BlockItemIds.ENCHANTING_TABLE,
@@ -2758,7 +_,7 @@
             .lightLevel(state -> 15)
             .strength(-1.0F, 3600000.0F)
             .noLootTable()
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block END_PORTAL_FRAME = register(
         BlockItemIds.END_PORTAL_FRAME,
@@ -2783,7 +_,7 @@
             .strength(3.0F, 9.0F)
             .lightLevel(state -> 1)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block REDSTONE_LAMP = register(
         BlockItemIds.REDSTONE_LAMP,
@@ -2804,18 +_,18 @@
             .strength(0.2F, 3.0F)
             .sound(SoundType.WOOD)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SHELF_MUSHROOM = register(
         BlockItemIds.SHELF_MUSHROOM,
         ShelfMushroomBlock::new,
         BlockBehaviour.Properties.of()
-            .mapColor(MapColor.TERRACOTTA_YELLOW)
+            .mapColor(MapColor.PLANT)
             .bounceRestitution(0.75F)
             .fallDistanceReduction(0.5F)
-            .sound(SoundType.SHELF_MUSHROOM)
+            .sound(SoundType.WOOD)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SANDSTONE_STAIRS = registerLegacyStair(BlockItemIds.SANDSTONE_STAIRS, SANDSTONE);
     public static final Block EMERALD_ORE = register(
@@ -2836,10 +_,10 @@
     public static final Block TRIPWIRE_HOOK = register(
         BlockItemIds.TRIPWIRE_HOOK,
         TripWireHookBlock::new,
-        BlockBehaviour.Properties.of().noCollision().sound(SoundType.WOOD).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().noCollision().sound(SoundType.WOOD).pushReaction(PushReaction.DESTROY)
     );
     public static final Block TRIPWIRE = register(
-        BlockItemIds.TRIPWIRE, p -> new TripWireBlock(TRIPWIRE_HOOK, p), BlockBehaviour.Properties.of().noCollision().pushReaction(PushReaction.POPPED)
+        BlockItemIds.TRIPWIRE, p -> new TripWireBlock(TRIPWIRE_HOOK, p), BlockBehaviour.Properties.of().noCollision().pushReaction(PushReaction.DESTROY)
     );
     public static final Block EMERALD_BLOCK = register(
         BlockItemIds.EMERALD_BLOCK,
@@ -2933,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.CROP)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block POTATOES = register(
         BlockItemIds.POTATO_CROP,
@@ -2944,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.CROP)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block OAK_BUTTON = register(BlockItemIds.OAK_BUTTON, p -> new ButtonBlock(BlockSetType.OAK, 30, p), buttonProperties());
     public static final Block SPRUCE_BUTTON = register(BlockItemIds.SPRUCE_BUTTON, p -> new ButtonBlock(BlockSetType.SPRUCE, 30, p), buttonProperties());
@@ -2960,68 +_,68 @@
     public static final Block SKELETON_SKULL = register(
         BlockItemIds.SKELETON_SKULL,
         p -> new SkullBlock(SkullBlock.Types.SKELETON, p),
-        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.SKELETON).strength(1.0F).pushReaction(PushReaction.POPPED).noOcclusion()
+        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.SKELETON).strength(1.0F).pushReaction(PushReaction.DESTROY).noOcclusion()
     );
     public static final Block SKELETON_WALL_SKULL = register(
         BlockIds.SKELETON_WALL_SKULL,
         p -> new WallSkullBlock(SkullBlock.Types.SKELETON, p),
-        wallVariant(SKELETON_SKULL, true).strength(1.0F).pushReaction(PushReaction.POPPED)
+        wallVariant(SKELETON_SKULL, true).strength(1.0F).pushReaction(PushReaction.DESTROY)
     );
     public static final Block WITHER_SKELETON_SKULL = register(
         BlockItemIds.WITHER_SKELETON_SKULL,
         WitherSkullBlock::new,
-        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.WITHER_SKELETON).strength(1.0F).pushReaction(PushReaction.POPPED).noOcclusion()
+        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.WITHER_SKELETON).strength(1.0F).pushReaction(PushReaction.DESTROY).noOcclusion()
     );
     public static final Block WITHER_SKELETON_WALL_SKULL = register(
         BlockIds.WITHER_SKELETON_WALL_SKULL,
         WitherWallSkullBlock::new,
-        wallVariant(WITHER_SKELETON_SKULL, true).strength(1.0F).pushReaction(PushReaction.POPPED)
+        wallVariant(WITHER_SKELETON_SKULL, true).strength(1.0F).pushReaction(PushReaction.DESTROY)
     );
     public static final Block ZOMBIE_HEAD = register(
         BlockItemIds.ZOMBIE_HEAD,
         p -> new SkullBlock(SkullBlock.Types.ZOMBIE, p),
-        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.ZOMBIE).strength(1.0F).pushReaction(PushReaction.POPPED).noOcclusion()
+        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.ZOMBIE).strength(1.0F).pushReaction(PushReaction.DESTROY).noOcclusion()
     );
     public static final Block ZOMBIE_WALL_HEAD = register(
         BlockIds.ZOMBIE_WALL_HEAD,
         p -> new WallSkullBlock(SkullBlock.Types.ZOMBIE, p),
-        wallVariant(ZOMBIE_HEAD, true).strength(1.0F).pushReaction(PushReaction.POPPED)
+        wallVariant(ZOMBIE_HEAD, true).strength(1.0F).pushReaction(PushReaction.DESTROY)
     );
     public static final Block PLAYER_HEAD = register(
         BlockItemIds.PLAYER_HEAD,
         PlayerHeadBlock::new,
-        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.CUSTOM_HEAD).strength(1.0F).pushReaction(PushReaction.POPPED).noOcclusion()
+        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.CUSTOM_HEAD).strength(1.0F).pushReaction(PushReaction.DESTROY).noOcclusion()
     );
     public static final Block PLAYER_WALL_HEAD = register(
-        BlockIds.PLAYER_WALL_HEAD, PlayerWallHeadBlock::new, wallVariant(PLAYER_HEAD, true).strength(1.0F).pushReaction(PushReaction.POPPED)
+        BlockIds.PLAYER_WALL_HEAD, PlayerWallHeadBlock::new, wallVariant(PLAYER_HEAD, true).strength(1.0F).pushReaction(PushReaction.DESTROY)
     );
     public static final Block CREEPER_HEAD = register(
         BlockItemIds.CREEPER_HEAD,
         p -> new SkullBlock(SkullBlock.Types.CREEPER, p),
-        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.CREEPER).strength(1.0F).pushReaction(PushReaction.POPPED).noOcclusion()
+        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.CREEPER).strength(1.0F).pushReaction(PushReaction.DESTROY).noOcclusion()
     );
     public static final Block CREEPER_WALL_HEAD = register(
         BlockIds.CREEPER_WALL_HEAD,
         p -> new WallSkullBlock(SkullBlock.Types.CREEPER, p),
-        wallVariant(CREEPER_HEAD, true).strength(1.0F).pushReaction(PushReaction.POPPED)
+        wallVariant(CREEPER_HEAD, true).strength(1.0F).pushReaction(PushReaction.DESTROY)
     );
     public static final Block DRAGON_HEAD = register(
         BlockItemIds.DRAGON_HEAD,
         p -> new SkullBlock(SkullBlock.Types.DRAGON, p),
-        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.DRAGON).strength(1.0F).pushReaction(PushReaction.POPPED).noOcclusion()
+        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.DRAGON).strength(1.0F).pushReaction(PushReaction.DESTROY).noOcclusion()
     );
     public static final Block DRAGON_WALL_HEAD = register(
         BlockIds.DRAGON_WALL_HEAD,
         p -> new WallSkullBlock(SkullBlock.Types.DRAGON, p),
-        wallVariant(DRAGON_HEAD, true).strength(1.0F).pushReaction(PushReaction.POPPED)
+        wallVariant(DRAGON_HEAD, true).strength(1.0F).pushReaction(PushReaction.DESTROY)
     );
     public static final Block PIGLIN_HEAD = register(
         BlockItemIds.PIGLIN_HEAD,
         p -> new SkullBlock(SkullBlock.Types.PIGLIN, p),
-        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.PIGLIN).strength(1.0F).pushReaction(PushReaction.POPPED).noOcclusion()
+        BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.PIGLIN).strength(1.0F).pushReaction(PushReaction.DESTROY).noOcclusion()
     );
     public static final Block PIGLIN_WALL_HEAD = register(
-        BlockIds.PIGLIN_WALL_HEAD, PiglinWallSkullBlock::new, wallVariant(PIGLIN_HEAD, true).strength(1.0F).pushReaction(PushReaction.POPPED)
+        BlockIds.PIGLIN_WALL_HEAD, PiglinWallSkullBlock::new, wallVariant(PIGLIN_HEAD, true).strength(1.0F).pushReaction(PushReaction.DESTROY)
     );
     public static final Block ANVIL = register(
         BlockItemIds.ANVIL,
@@ -3031,7 +_,7 @@
             .requiresCorrectToolForDrops()
             .strength(5.0F, 1200.0F)
             .sound(SoundType.ANVIL)
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block CHIPPED_ANVIL = register(
         BlockItemIds.CHIPPED_ANVIL,
@@ -3041,7 +_,7 @@
             .requiresCorrectToolForDrops()
             .strength(5.0F, 1200.0F)
             .sound(SoundType.ANVIL)
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block DAMAGED_ANVIL = register(
         BlockItemIds.DAMAGED_ANVIL,
@@ -3051,7 +_,7 @@
             .requiresCorrectToolForDrops()
             .strength(5.0F, 1200.0F)
             .sound(SoundType.ANVIL)
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block TRAPPED_CHEST = register(
         BlockItemIds.TRAPPED_CHEST,
@@ -3061,15 +_,15 @@
     public static final Block LIGHT_WEIGHTED_PRESSURE_PLATE = register(
         BlockItemIds.LIGHT_WEIGHTED_PRESSURE_PLATE,
         p -> new WeightedPressurePlateBlock(15, BlockSetType.GOLD, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.GOLD).forceSolidOn().noCollision().strength(0.5F).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.GOLD).forceSolidOn().noCollision().strength(0.5F).pushReaction(PushReaction.DESTROY)
     );
     public static final Block HEAVY_WEIGHTED_PRESSURE_PLATE = register(
         BlockItemIds.HEAVY_WEIGHTED_PRESSURE_PLATE,
         p -> new WeightedPressurePlateBlock(150, BlockSetType.IRON, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.METAL).forceSolidOn().noCollision().strength(0.5F).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.METAL).forceSolidOn().noCollision().strength(0.5F).pushReaction(PushReaction.DESTROY)
     );
     public static final Block COMPARATOR = register(
-        BlockItemIds.COMPARATOR, ComparatorBlock::new, BlockBehaviour.Properties.of().instabreak().sound(SoundType.STONE).pushReaction(PushReaction.POPPED)
+        BlockItemIds.COMPARATOR, ComparatorBlock::new, BlockBehaviour.Properties.of().instabreak().sound(SoundType.STONE).pushReaction(PushReaction.DESTROY)
     );
     public static final Block DAYLIGHT_DETECTOR = register(
         BlockItemIds.DAYLIGHT_DETECTOR,
@@ -3162,7 +_,7 @@
             .noOcclusion()
             .isValidSpawn(Blocks::never)
             .noTerrainParticles()
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block LIGHT = register(
         BlockItemIds.LIGHT,
@@ -3253,7 +_,7 @@
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block LILAC = register(
         BlockItemIds.LILAC,
@@ -3265,7 +_,7 @@
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block ROSE_BUSH = register(
         BlockItemIds.ROSE_BUSH,
@@ -3277,7 +_,7 @@
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block PEONY = register(
         BlockItemIds.PEONY,
@@ -3289,7 +_,7 @@
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block TALL_GRASS = register(
         BlockItemIds.TALL_GRASS,
@@ -3302,7 +_,7 @@
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block LARGE_FERN = register(
         BlockItemIds.LARGE_FERN,
@@ -3315,7 +_,7 @@
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final ColorCollection<Block> BANNER = ColorCollection.registerBlocks(
         BlockItemIds.BANNER,
@@ -3614,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BIRCH_DOOR = register(
         BlockItemIds.BIRCH_DOOR,
@@ -3625,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block JUNGLE_DOOR = register(
         BlockItemIds.JUNGLE_DOOR,
@@ -3636,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block ACACIA_DOOR = register(
         BlockItemIds.ACACIA_DOOR,
@@ -3647,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CHERRY_DOOR = register(
         BlockItemIds.CHERRY_DOOR,
@@ -3658,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block DARK_OAK_DOOR = register(
         BlockItemIds.DARK_OAK_DOOR,
@@ -3669,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block PALE_OAK_DOOR = register(
         BlockItemIds.PALE_OAK_DOOR,
@@ -3680,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block MANGROVE_DOOR = register(
         BlockItemIds.MANGROVE_DOOR,
@@ -3691,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block POPLAR_DOOR = register(
         BlockItemIds.POPLAR_DOOR,
@@ -3702,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BAMBOO_DOOR = register(
         BlockItemIds.BAMBOO_DOOR,
@@ -3713,7 +_,7 @@
             .strength(3.0F)
             .noOcclusion()
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block END_ROD = register(
         BlockItemIds.END_ROD,
@@ -3729,7 +_,7 @@
             .strength(0.4F)
             .sound(SoundType.WOOD)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CHORUS_FLOWER = register(
         BlockItemIds.CHORUS_FLOWER,
@@ -3742,7 +_,7 @@
             .sound(SoundType.WOOD)
             .noOcclusion()
             .isValidSpawn(Blocks::never)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .isRedstoneConductor(Blocks::never)
     );
     public static final Block PURPUR_BLOCK = register(
@@ -3777,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.CROP)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block PITCHER_CROP = register(
         BlockItemIds.PITCHER_CROP,
@@ -3788,7 +_,7 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.CROP)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block PITCHER_PLANT = register(
         BlockItemIds.PITCHER_PLANT,
@@ -3800,7 +_,7 @@
             .sound(SoundType.CROP)
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BEETROOTS = register(
         BlockItemIds.BEETROOT_CROP,
@@ -3811,12 +_,17 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.CROP)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block DIRT_PATH = register(
         BlockItemIds.DIRT_PATH,
         p -> new PathBlock(DIRT, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.DIRT).strength(0.65F).sound(SoundType.GRASS).isSuffocating(Blocks::always)
+        BlockBehaviour.Properties.of()
+            .mapColor(MapColor.DIRT)
+            .strength(0.65F)
+            .sound(SoundType.GRASS)
+            .isViewBlocking(Blocks::always)
+            .isSuffocating(Blocks::always)
     );
     public static final Block END_GATEWAY = register(
         BlockIds.END_GATEWAY,
@@ -3827,7 +_,7 @@
             .lightLevel(state -> 15)
             .strength(-1.0F, 3600000.0F)
             .noLootTable()
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block REPEATING_COMMAND_BLOCK = register(
         BlockItemIds.REPEATING_COMMAND_BLOCK,
@@ -3889,7 +_,7 @@
     public static final Block STRUCTURE_VOID = register(
         BlockItemIds.STRUCTURE_VOID,
         StructureVoidBlock::new,
-        BlockBehaviour.Properties.of().replaceable().noCollision().noLootTable().noTerrainParticles().pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().replaceable().noCollision().noLootTable().noTerrainParticles().pushReaction(PushReaction.DESTROY)
     );
     public static final Block OBSERVER = register(
         BlockItemIds.OBSERVER,
@@ -3917,7 +_,7 @@
             .instrument(NoteBlockInstrument.BASEDRUM)
             .requiresCorrectToolForDrops()
             .strength(1.4F)
-            .pushReaction(PushReaction.PUSH)
+            .pushReaction(PushReaction.PUSH_ONLY)
     );
     public static final ColorCollection<Block> CONCRETE = ColorCollection.registerBlocks(
         BlockItemIds.CONCRETE,
@@ -3929,12 +_,6 @@
             .requiresCorrectToolForDrops()
             .strength(1.8F)
     );
-    public static final ColorCollection<Block> CONCRETE_STAIRS = ColorCollection.zipMap(
-        ColorCollection.VALUES, BlockItemIds.CONCRETE_STAIRS, (color, id) -> registerStair(id, CONCRETE.pick(color))
-    );
-    public static final ColorCollection<Block> CONCRETE_SLAB = ColorCollection.zipMap(
-        ColorCollection.VALUES, BlockItemIds.CONCRETE_SLAB, (color, id) -> registerSlab(id, CONCRETE.pick(color))
-    );
     public static final ColorCollection<Block> CONCRETE_POWDER = ColorCollection.registerBlocks(
         BlockItemIds.CONCRETE_POWDER,
         Blocks::register,
@@ -3950,12 +_,12 @@
             .randomTicks()
             .instabreak()
             .sound(SoundType.WET_GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block KELP_PLANT = register(
         BlockIds.KELP_PLANT,
         KelpPlantBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.WATER).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.WATER).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block DRIED_KELP_BLOCK = register(
         BlockItemIds.DRIED_KELP_BLOCK, BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_GREEN).strength(0.5F, 2.5F).sound(SoundType.GRASS)
@@ -3970,7 +_,7 @@
             .sound(SoundType.METAL)
             .randomTicks()
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SNIFFER_EGG = register(
         BlockItemIds.SNIFFER_EGG,
@@ -4135,27 +_,27 @@
     public static final Block TUBE_CORAL = register(
         BlockItemIds.TUBE_CORAL,
         p -> new CoralPlantBlock(DEAD_TUBE_CORAL, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_BLUE).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_BLUE).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block BRAIN_CORAL = register(
         BlockItemIds.BRAIN_CORAL,
         p -> new CoralPlantBlock(DEAD_BRAIN_CORAL, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_PINK).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_PINK).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block BUBBLE_CORAL = register(
         BlockItemIds.BUBBLE_CORAL,
         p -> new CoralPlantBlock(DEAD_BUBBLE_CORAL, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_PURPLE).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_PURPLE).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block FIRE_CORAL = register(
         BlockItemIds.FIRE_CORAL,
         p -> new CoralPlantBlock(DEAD_FIRE_CORAL, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_RED).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_RED).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block HORN_CORAL = register(
         BlockItemIds.HORN_CORAL,
         p -> new CoralPlantBlock(DEAD_HORN_CORAL, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_YELLOW).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_YELLOW).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block DEAD_TUBE_CORAL_FAN = register(
         BlockItemIds.DEAD_TUBE_CORAL_FAN,
@@ -4215,27 +_,27 @@
     public static final Block TUBE_CORAL_FAN = register(
         BlockItemIds.TUBE_CORAL_FAN,
         p -> new CoralFanBlock(DEAD_TUBE_CORAL_FAN, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_BLUE).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_BLUE).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block BRAIN_CORAL_FAN = register(
         BlockItemIds.BRAIN_CORAL_FAN,
         p -> new CoralFanBlock(DEAD_BRAIN_CORAL_FAN, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_PINK).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_PINK).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block BUBBLE_CORAL_FAN = register(
         BlockItemIds.BUBBLE_CORAL_FAN,
         p -> new CoralFanBlock(DEAD_BUBBLE_CORAL_FAN, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_PURPLE).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_PURPLE).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block FIRE_CORAL_FAN = register(
         BlockItemIds.FIRE_CORAL_FAN,
         p -> new CoralFanBlock(DEAD_FIRE_CORAL_FAN, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_RED).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_RED).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block HORN_CORAL_FAN = register(
         BlockItemIds.HORN_CORAL_FAN,
         p -> new CoralFanBlock(DEAD_HORN_CORAL_FAN, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_YELLOW).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_YELLOW).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block DEAD_TUBE_CORAL_WALL_FAN = register(
         BlockIds.DEAD_TUBE_CORAL_WALL_FAN,
@@ -4300,7 +_,7 @@
             .noCollision()
             .instabreak()
             .sound(SoundType.WET_GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BRAIN_CORAL_WALL_FAN = register(
         BlockIds.BRAIN_CORAL_WALL_FAN,
@@ -4310,7 +_,7 @@
             .noCollision()
             .instabreak()
             .sound(SoundType.WET_GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BUBBLE_CORAL_WALL_FAN = register(
         BlockIds.BUBBLE_CORAL_WALL_FAN,
@@ -4320,12 +_,17 @@
             .noCollision()
             .instabreak()
             .sound(SoundType.WET_GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block FIRE_CORAL_WALL_FAN = register(
         BlockIds.FIRE_CORAL_WALL_FAN,
         p -> new CoralWallFanBlock(DEAD_FIRE_CORAL_WALL_FAN, p),
-        wallVariant(FIRE_CORAL_FAN, false).mapColor(MapColor.COLOR_RED).noCollision().instabreak().sound(SoundType.WET_GRASS).pushReaction(PushReaction.POPPED)
+        wallVariant(FIRE_CORAL_FAN, false)
+            .mapColor(MapColor.COLOR_RED)
+            .noCollision()
+            .instabreak()
+            .sound(SoundType.WET_GRASS)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block HORN_CORAL_WALL_FAN = register(
         BlockIds.HORN_CORAL_WALL_FAN,
@@ -4335,7 +_,7 @@
             .noCollision()
             .instabreak()
             .sound(SoundType.WET_GRASS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SEA_PICKLE = register(
         BlockItemIds.SEA_PICKLE,
@@ -4345,7 +_,7 @@
             .lightLevel(state -> SeaPickleBlock.isDead(state) ? 0 : 3 + 3 * state.getValue(SeaPickleBlock.PICKLES))
             .sound(SoundType.SLIME_BLOCK)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BLUE_ICE = register(
         BlockItemIds.BLUE_ICE,
@@ -4376,7 +_,7 @@
             .sound(SoundType.BAMBOO_SAPLING)
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block BAMBOO = register(
         BlockItemIds.BAMBOO,
@@ -4392,7 +_,7 @@
             .dynamicShape()
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .isRedstoneConductor(Blocks::never)
     );
     public static final Block POTTED_BAMBOO = register(BlockIds.POTTED_BAMBOO, p -> new FlowerPotBlock(BAMBOO, p), flowerPotProperties());
@@ -4410,7 +_,7 @@
             .replaceable()
             .noCollision()
             .noLootTable()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .liquid()
             .sound(SoundType.EMPTY)
     );
@@ -4463,7 +_,7 @@
             .sound(SoundType.SCAFFOLDING)
             .dynamicShape()
             .isValidSpawn(Blocks::never)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .isRedstoneConductor(Blocks::never)
     );
     public static final Block LOOM = register(
@@ -4513,7 +_,7 @@
             .requiresCorrectToolForDrops()
             .strength(2.0F, 6.0F)
             .sound(SoundType.STONE)
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block LECTERN = register(
         BlockItemIds.LECTERN,
@@ -4533,7 +_,7 @@
     public static final Block BELL = register(
         BlockItemIds.BELL,
         BellBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.GOLD).forceSolidOn().strength(5.0F).sound(SoundType.ANVIL).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.GOLD).forceSolidOn().strength(5.0F).sound(SoundType.ANVIL).pushReaction(PushReaction.DESTROY)
     );
     public static final Block LANTERN = register(
         BlockItemIds.LANTERN,
@@ -4545,7 +_,7 @@
             .sound(SoundType.LANTERN)
             .lightLevel(state -> 15)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SOUL_LANTERN = register(
         BlockItemIds.SOUL_LANTERN,
@@ -4557,7 +_,7 @@
             .sound(SoundType.LANTERN)
             .lightLevel(state -> 10)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final WeatheringCopperCollection<Block> COPPER_LANTERN = WeatheringCopperCollection.registerBlocks(
         BlockItemIds.COPPER_LANTERN,
@@ -4571,7 +_,7 @@
             .sound(SoundType.LANTERN)
             .lightLevel(state -> 15)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CAMPFIRE = register(
         BlockItemIds.CAMPFIRE,
@@ -4600,7 +_,12 @@
     public static final Block SWEET_BERRY_BUSH = register(
         BlockItemIds.SWEET_BERRY_CROP,
         SweetBerryBushBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).randomTicks().noCollision().sound(SoundType.SWEET_BERRY_BUSH).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of()
+            .mapColor(MapColor.PLANT)
+            .randomTicks()
+            .noCollision()
+            .sound(SoundType.SWEET_BERRY_BUSH)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block WARPED_STEM = register(BlockItemIds.WARPED_STEM, RotatedPillarBlock::new, netherStemProperties(MapColor.WARPED_STEM));
     public static final Block STRIPPED_WARPED_STEM = register(
@@ -4630,7 +_,7 @@
     public static final Block WARPED_FUNGUS = register(
         BlockItemIds.WARPED_FUNGUS,
         p -> new NetherFungusBlock(TreeFeatures.WARPED_FUNGUS_PLANTED, WARPED_NYLIUM, BlockTags.SUPPORTS_WARPED_FUNGUS, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_CYAN).instabreak().noCollision().sound(SoundType.FUNGUS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_CYAN).instabreak().noCollision().sound(SoundType.FUNGUS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block WARPED_WART_BLOCK = register(
         BlockItemIds.WARPED_WART_BLOCK, BlockBehaviour.Properties.of().mapColor(MapColor.WARPED_WART_BLOCK).strength(1.0F).sound(SoundType.WART_BLOCK)
@@ -4645,7 +_,7 @@
             .instabreak()
             .sound(SoundType.ROOTS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block NETHER_SPROUTS = register(
         BlockItemIds.NETHER_SPROUTS,
@@ -4657,7 +_,7 @@
             .instabreak()
             .sound(SoundType.NETHER_SPROUTS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CRIMSON_STEM = register(BlockItemIds.CRIMSON_STEM, RotatedPillarBlock::new, netherStemProperties(MapColor.CRIMSON_STEM));
     public static final Block STRIPPED_CRIMSON_STEM = register(
@@ -4687,7 +_,7 @@
     public static final Block CRIMSON_FUNGUS = register(
         BlockItemIds.CRIMSON_FUNGUS,
         p -> new NetherFungusBlock(TreeFeatures.CRIMSON_FUNGUS_PLANTED, CRIMSON_NYLIUM, BlockTags.SUPPORTS_CRIMSON_FUNGUS, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.NETHER).instabreak().noCollision().sound(SoundType.FUNGUS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.NETHER).instabreak().noCollision().sound(SoundType.FUNGUS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block SHROOMLIGHT = register(
         BlockItemIds.SHROOMLIGHT,
@@ -4702,12 +_,12 @@
             .noCollision()
             .instabreak()
             .sound(SoundType.WEEPING_VINES)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block WEEPING_VINES_PLANT = register(
         BlockIds.WEEPING_VINES_PLANT,
         WeepingVinesPlantBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.NETHER).noCollision().instabreak().sound(SoundType.WEEPING_VINES).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.NETHER).noCollision().instabreak().sound(SoundType.WEEPING_VINES).pushReaction(PushReaction.DESTROY)
     );
     public static final Block TWISTING_VINES = register(
         BlockItemIds.TWISTING_VINES,
@@ -4718,7 +_,7 @@
             .noCollision()
             .instabreak()
             .sound(SoundType.WEEPING_VINES)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block TWISTING_VINES_PLANT = register(
         BlockIds.TWISTING_VINES_PLANT,
@@ -4728,7 +_,7 @@
             .noCollision()
             .instabreak()
             .sound(SoundType.WEEPING_VINES)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CRIMSON_ROOTS = register(
         BlockItemIds.CRIMSON_ROOTS,
@@ -4740,7 +_,7 @@
             .instabreak()
             .sound(SoundType.ROOTS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CRIMSON_PLANKS = register(
         BlockItemIds.CRIMSON_PLANKS,
@@ -4761,7 +_,7 @@
             .instrument(NoteBlockInstrument.BASS)
             .noCollision()
             .strength(0.5F)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block WARPED_PRESSURE_PLATE = register(
         BlockItemIds.WARPED_PRESSURE_PLATE,
@@ -4772,7 +_,7 @@
             .instrument(NoteBlockInstrument.BASS)
             .noCollision()
             .strength(0.5F)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CRIMSON_FENCE = register(
         BlockItemIds.CRIMSON_FENCE,
@@ -4834,7 +_,7 @@
             .instrument(NoteBlockInstrument.BASS)
             .strength(3.0F)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block WARPED_DOOR = register(
         BlockItemIds.WARPED_DOOR,
@@ -4844,7 +_,7 @@
             .instrument(NoteBlockInstrument.BASS)
             .strength(3.0F)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CRIMSON_SIGN = register(
         BlockItemIds.CRIMSON_SIGN,
@@ -4900,7 +_,9 @@
         BlockItemIds.TEST_BLOCK, TestBlock::new, BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_LIGHT_GRAY).strength(-1.0F, 3600000.0F).noLootTable()
     );
     public static final Block TEST_INSTANCE_BLOCK = register(
-        BlockItemIds.TEST_INSTANCE_BLOCK, TestInstanceBlock::new, BlockBehaviour.Properties.of().noOcclusion().strength(-1.0F, 3600000.0F).noLootTable()
+        BlockItemIds.TEST_INSTANCE_BLOCK,
+        TestInstanceBlock::new,
+        BlockBehaviour.Properties.of().noOcclusion().strength(-1.0F, 3600000.0F).noLootTable().isViewBlocking(Blocks::never)
     );
     public static final Block COMPOSTER = register(
         BlockItemIds.COMPOSTER,
@@ -4950,7 +_,6 @@
             .requiresCorrectToolForDrops()
             .strength(50.0F, 1200.0F)
             .lightLevel(state -> 10)
-            .pushReaction(PushReaction.IMMOVEABLE)
     );
     public static final Block RESPAWN_ANCHOR = register(
         BlockItemIds.RESPAWN_ANCHOR,
@@ -4961,7 +_,6 @@
             .requiresCorrectToolForDrops()
             .strength(50.0F, 1200.0F)
             .lightLevel(state -> RespawnAnchorBlock.getScaledChargeLevel(state, 15))
-            .pushReaction(PushReaction.IMMOVEABLE)
     );
     public static final Block POTTED_CRIMSON_FUNGUS = register(
         BlockIds.POTTED_CRIMSON_FUNGUS, p -> new FlowerPotBlock(CRIMSON_FUNGUS, p), flowerPotProperties()
@@ -4976,7 +_,7 @@
             .requiresCorrectToolForDrops()
             .strength(3.5F)
             .sound(SoundType.LODESTONE)
-            .pushReaction(PushReaction.IMMOVEABLE)
+            .pushReaction(PushReaction.BLOCK)
     );
     public static final Block BLACKSTONE = register(
         BlockItemIds.BLACKSTONE,
@@ -5018,7 +_,7 @@
             .instrument(NoteBlockInstrument.BASEDRUM)
             .noCollision()
             .strength(0.5F)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block POLISHED_BLACKSTONE_BUTTON = register(
         BlockItemIds.POLISHED_BLACKSTONE_BUTTON, p -> new ButtonBlock(BlockSetType.STONE, 20, p), buttonProperties()
@@ -5073,7 +_,7 @@
             .strength(1.5F)
             .sound(SoundType.AMETHYST)
             .requiresCorrectToolForDrops()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block AMETHYST_CLUSTER = register(
         BlockItemIds.AMETHYST_CLUSTER,
@@ -5085,7 +_,7 @@
             .sound(SoundType.AMETHYST_CLUSTER)
             .strength(1.5F)
             .lightLevel(state -> 5)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block LARGE_AMETHYST_BUD = register(
         BlockItemIds.LARGE_AMETHYST_BUD,
@@ -5187,6 +_,7 @@
             .isValidSpawn(Blocks::never)
             .isRedstoneConductor(Blocks::never)
             .isSuffocating(Blocks::never)
+            .isViewBlocking(Blocks::never)
     );
     public static final Block POWDER_SNOW = register(
         BlockItemIds.POWDER_SNOW,
@@ -5224,7 +_,7 @@
             .noCollision()
             .strength(0.2F)
             .sound(SoundType.SCULK_VEIN)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SCULK_CATALYST = register(
         BlockItemIds.SCULK_CATALYST,
@@ -5281,14 +_,14 @@
         Blocks::register,
         (state, p) -> new StairBlock(CUT_COPPER.waxed().pick(state).defaultBlockState(), p),
         (state, p) -> new WeatheringCopperStairBlock(state, CUT_COPPER.weathering().pick(state).defaultBlockState(), p),
-        state -> BlockBehaviour.Properties.ofFullCopy(CUT_COPPER.weathering().pick(state)).isViewBlocking(NEAR_PLANE_INTERSECTS_OUTLINE)
+        state -> BlockBehaviour.Properties.ofFullCopy(CUT_COPPER.weathering().pick(state))
     );
     public static final WeatheringCopperCollection<Block> CUT_COPPER_SLAB = WeatheringCopperCollection.registerBlocks(
         BlockItemIds.CUT_COPPER_SLAB,
         Blocks::register,
         (var0, p) -> new SlabBlock(p),
         WeatheringCopperSlabBlock::new,
-        state -> BlockBehaviour.Properties.ofFullCopy(CUT_COPPER.weathering().pick(state)).isViewBlocking(NEAR_PLANE_INTERSECTS_OUTLINE)
+        state -> BlockBehaviour.Properties.ofFullCopy(CUT_COPPER.weathering().pick(state))
     );
     public static final WeatheringCopperCollection<Block> COPPER_DOOR = WeatheringCopperCollection.registerBlocks(
         BlockItemIds.COPPER_DOOR,
@@ -5299,7 +_,7 @@
             .mapColor(var1 -> COPPER_BLOCK.weathering().pick(state).defaultMapColor())
             .strength(3.0F, 6.0F)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final WeatheringCopperCollection<Block> COPPER_TRAPDOOR = WeatheringCopperCollection.registerBlocks(
         BlockItemIds.COPPER_TRAPDOOR,
@@ -5327,6 +_,7 @@
             .isValidSpawn(Blocks::never)
             .isRedstoneConductor(Blocks::never)
             .isSuffocating(Blocks::never)
+            .isViewBlocking(Blocks::never)
     );
     public static final WeatheringCopperCollection<Block> COPPER_BULB = WeatheringCopperCollection.registerBlocks(
         BlockItemIds.COPPER_BULB,
@@ -5368,7 +_,7 @@
             .mapColor(var1 -> COPPER_BLOCK.weathering().pick(state).defaultMapColor())
             .strength(3.0F, 6.0F)
             .sound(SoundType.COPPER_GOLEM_STATUE)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .noOcclusion()
     );
     public static final WeatheringCopperCollection<Block> LIGHTNING_ROD = WeatheringCopperCollection.registerBlocks(
@@ -5406,7 +_,7 @@
             .strength(1.5F, 3.0F)
             .dynamicShape()
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .isRedstoneConductor(Blocks::never)
             .noOcclusion()
     );
@@ -5423,7 +_,7 @@
             .strength(1.5F, 3.0F)
             .dynamicShape()
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .isRedstoneConductor(Blocks::never)
             .noOcclusion()
     );
@@ -5437,7 +_,7 @@
             .lightLevel(CaveVines.emission(14))
             .instabreak()
             .sound(SoundType.CAVE_VINES)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block CAVE_VINES_PLANT = register(
         BlockIds.CAVE_VINES_PLANT,
@@ -5448,12 +_,12 @@
             .lightLevel(CaveVines.emission(14))
             .instabreak()
             .sound(SoundType.CAVE_VINES)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block SPORE_BLOSSOM = register(
         BlockItemIds.SPORE_BLOSSOM,
         SporeBlossomBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).instabreak().noCollision().sound(SoundType.SPORE_BLOSSOM).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).instabreak().noCollision().sound(SoundType.SPORE_BLOSSOM).pushReaction(PushReaction.DESTROY)
     );
     public static final Block AZALEA = register(
         BlockItemIds.AZALEA,
@@ -5464,7 +_,7 @@
             .instabreak()
             .sound(SoundType.AZALEA)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block FLOWERING_AZALEA = register(
         BlockItemIds.FLOWERING_AZALEA,
@@ -5475,22 +_,22 @@
             .instabreak()
             .sound(SoundType.FLOWERING_AZALEA)
             .noOcclusion()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block MOSS_CARPET = register(
         BlockItemIds.MOSS_CARPET,
         CarpetBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_GREEN).strength(0.1F).sound(SoundType.MOSS_CARPET).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_GREEN).strength(0.1F).sound(SoundType.MOSS_CARPET).pushReaction(PushReaction.DESTROY)
     );
     public static final Block PINK_PETALS = register(
         BlockItemIds.PINK_PETALS,
         p -> new FlowerBedBlock(p, 3),
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().sound(SoundType.PINK_PETALS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().sound(SoundType.PINK_PETALS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block WILDFLOWERS = register(
         BlockItemIds.WILDFLOWERS,
         p -> new FlowerBedBlock(p, 3),
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().sound(SoundType.PINK_PETALS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().sound(SoundType.PINK_PETALS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block LEAF_LITTER = register(
         BlockItemIds.LEAF_LITTER,
@@ -5500,22 +_,22 @@
             .replaceable()
             .noCollision()
             .sound(SoundType.LEAF_LITTER)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block MOSS_BLOCK = register(
         BlockItemIds.MOSS_BLOCK,
         p -> new BonemealableFeaturePlacerBlock(CaveFeatures.MOSS_PATCH_BONEMEAL, p),
-        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_GREEN).strength(0.1F).sound(SoundType.MOSS).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.COLOR_GREEN).strength(0.1F).sound(SoundType.MOSS).pushReaction(PushReaction.DESTROY)
     );
     public static final Block BIG_DRIPLEAF = register(
         BlockItemIds.BIG_DRIPLEAF,
         BigDripleafBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).forceSolidOff().strength(0.1F).sound(SoundType.BIG_DRIPLEAF).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).forceSolidOff().strength(0.1F).sound(SoundType.BIG_DRIPLEAF).pushReaction(PushReaction.DESTROY)
     );
     public static final Block BIG_DRIPLEAF_STEM = register(
         BlockIds.BIG_DRIPLEAF_STEM,
         BigDripleafStemBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().strength(0.1F).sound(SoundType.BIG_DRIPLEAF).pushReaction(PushReaction.POPPED)
+        BlockBehaviour.Properties.of().mapColor(MapColor.PLANT).noCollision().strength(0.1F).sound(SoundType.BIG_DRIPLEAF).pushReaction(PushReaction.DESTROY)
     );
     public static final Block SMALL_DRIPLEAF = register(
         BlockItemIds.SMALL_DRIPLEAF,
@@ -5526,7 +_,7 @@
             .instabreak()
             .sound(SoundType.SMALL_DRIPLEAF)
             .offsetType(BlockBehaviour.OffsetType.XYZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block HANGING_ROOTS = register(
         BlockItemIds.HANGING_ROOTS,
@@ -5539,7 +_,7 @@
             .sound(SoundType.HANGING_ROOTS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block ROOTED_DIRT = register(
         BlockItemIds.ROOTED_DIRT, RootedDirtBlock::new, BlockBehaviour.Properties.of().mapColor(MapColor.DIRT).strength(0.5F).sound(SoundType.ROOTED_DIRT)
@@ -5551,6 +_,7 @@
             .mapColor(MapColor.TERRACOTTA_CYAN)
             .isValidSpawn(Blocks::always)
             .isRedstoneConductor(Blocks::always)
+            .isViewBlocking(Blocks::always)
             .isSuffocating(Blocks::always)
             .sound(SoundType.MUD)
     );
@@ -5645,7 +_,7 @@
             .noOcclusion()
             .noCollision()
             .sound(SoundType.FROGSPAWN)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block REINFORCED_DEEPSLATE = register(
         BlockItemIds.REINFORCED_DEEPSLATE,
@@ -5654,12 +_,11 @@
             .instrument(NoteBlockInstrument.BASEDRUM)
             .sound(SoundType.DEEPSLATE)
             .strength(55.0F, 1200.0F)
-            .pushReaction(PushReaction.IMMOVEABLE)
     );
     public static final Block DECORATED_POT = register(
         BlockItemIds.DECORATED_POT,
         DecoratedPotBlock::new,
-        BlockBehaviour.Properties.of().mapColor(MapColor.TERRACOTTA_RED).strength(0.0F, 0.0F).pushReaction(PushReaction.POPPED).noOcclusion()
+        BlockBehaviour.Properties.of().mapColor(MapColor.TERRACOTTA_RED).strength(0.0F, 0.0F).pushReaction(PushReaction.DESTROY).noOcclusion()
     );
     public static final Block CRAFTER = register(
         BlockItemIds.CRAFTER, CrafterBlock::new, BlockBehaviour.Properties.of().mapColor(MapColor.STONE).strength(1.5F, 3.5F)
@@ -5673,6 +_,7 @@
             .lightLevel(state -> state.getValue(TrialSpawnerBlock.STATE).lightLevel())
             .strength(50.0F)
             .sound(SoundType.TRIAL_SPAWNER)
+            .isViewBlocking(Blocks::never)
             .noOcclusion()
     );
     public static final Block VAULT = register(
@@ -5685,6 +_,7 @@
             .sound(SoundType.VAULT)
             .lightLevel(state -> state.getValue(VaultBlock.STATE).lightLevel())
             .strength(50.0F)
+            .isViewBlocking(Blocks::never)
     );
     public static final Block HEAVY_CORE = register(
         BlockItemIds.HEAVY_CORE,
@@ -5694,7 +_,7 @@
             .instrument(NoteBlockInstrument.SNARE)
             .sound(SoundType.HEAVY_CORE)
             .strength(10.0F)
-            .pushReaction(PushReaction.PUSH_PULL)
+            .pushReaction(PushReaction.NORMAL)
             .explosionResistance(1200.0F)
     );
     public static final Block PALE_MOSS_BLOCK = register(
@@ -5705,7 +_,7 @@
             .mapColor(MapColor.COLOR_LIGHT_GRAY)
             .strength(0.1F)
             .sound(SoundType.MOSS)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block PALE_MOSS_CARPET = register(
         BlockItemIds.PALE_MOSS_CARPET,
@@ -5715,7 +_,7 @@
             .mapColor(PALE_MOSS_BLOCK.defaultMapColor())
             .strength(0.1F)
             .sound(SoundType.MOSS_CARPET)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .noOcclusion()
     );
     public static final Block PALE_HANGING_MOSS = register(
@@ -5726,7 +_,7 @@
             .mapColor(PALE_MOSS_BLOCK.defaultMapColor())
             .noCollision()
             .sound(SoundType.MOSS_CARPET)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
     public static final Block OPEN_EYEBLOSSOM = register(
         BlockItemIds.OPEN_EYEBLOSSOM,
@@ -5737,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .randomTicks()
     );
     public static final Block CLOSED_EYEBLOSSOM = register(
@@ -5749,7 +_,7 @@
             .instabreak()
             .sound(SoundType.GRASS)
             .offsetType(BlockBehaviour.OffsetType.XZ)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .randomTicks()
     );
     public static final Block POTTED_OPEN_EYEBLOSSOM = register(
@@ -5768,7 +_,7 @@
             .noCollision()
             .instabreak()
             .sound(SoundType.SWEET_BERRY_BUSH)
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
     );
 
     private static ToIntFunction<BlockState> litBlockEmission(int lightEmission) {
@@ -5779,15 +_,15 @@
         return blockState -> blockState.getValue(BlockStateProperties.WATERLOGGED) ? MapColor.WATER : mapColor;
     }
 
-    public static Boolean never(BlockState state, BlockGetter blockGetter, BlockPos blockPos, EntityType<?> entityType) {
+    private static Boolean never(BlockState state, BlockGetter blockGetter, BlockPos blockPos, EntityType<?> entityType) {
         return false;
     }
 
-    public static Boolean always(BlockState state, BlockGetter blockGetter, BlockPos blockPos, EntityType<?> entityType) {
+    private static Boolean always(BlockState state, BlockGetter blockGetter, BlockPos blockPos, EntityType<?> entityType) {
         return true;
     }
 
-    public static Boolean ocelotOrParrot(BlockState state, BlockGetter blockGetter, BlockPos blockPos, EntityType<?> entityType) {
+    private static Boolean ocelotOrParrot(BlockState state, BlockGetter blockGetter, BlockPos blockPos, EntityType<?> entityType) {
         return entityType == EntityTypes.OCELOT || entityType == EntityTypes.PARROT;
     }
 
@@ -5829,8 +_,9 @@
             .noOcclusion()
             .isValidSpawn(Blocks::ocelotOrParrot)
             .isSuffocating(Blocks::never)
+            .isViewBlocking(Blocks::never)
             .ignitedByLava()
-            .pushReaction(PushReaction.POPPED)
+            .pushReaction(PushReaction.DESTROY)
             .isRedstoneConductor(Blocks::never);
     }
 
@@ -5842,7 +_,8 @@
             .dynamicShape()
             .noOcclusion()
             .isSuffocating(NOT_CLOSED_SHULKER)
-            .pushReaction(PushReaction.POPPED);
+            .isViewBlocking(NOT_CLOSED_SHULKER)
+            .pushReaction(PushReaction.DESTROY);
     }
 
     private static BlockBehaviour.Properties pistonProperties() {
@@ -5851,15 +_,16 @@
             .strength(1.5F)
             .isRedstoneConductor(Blocks::never)
             .isSuffocating(NOT_EXTENDED_PISTON)
-            .pushReaction(PushReaction.IMMOVEABLE);
+            .isViewBlocking(NOT_EXTENDED_PISTON)
+            .pushReaction(PushReaction.BLOCK);
     }
 
     private static BlockBehaviour.Properties buttonProperties() {
-        return BlockBehaviour.Properties.of().noCollision().strength(0.5F).pushReaction(PushReaction.POPPED);
+        return BlockBehaviour.Properties.of().noCollision().strength(0.5F).pushReaction(PushReaction.DESTROY);
     }
 
     private static BlockBehaviour.Properties flowerPotProperties() {
-        return BlockBehaviour.Properties.of().instabreak().noOcclusion().pushReaction(PushReaction.POPPED);
+        return BlockBehaviour.Properties.of().instabreak().noOcclusion().pushReaction(PushReaction.DESTROY);
     }
 
     private static BlockBehaviour.Properties candleProperties(MapColor color) {
@@ -5869,37 +_,27 @@
             .strength(0.1F)
             .sound(SoundType.CANDLE)
             .lightLevel(CandleBlock.LIGHT_EMISSION)
-            .pushReaction(PushReaction.POPPED);
+            .pushReaction(PushReaction.DESTROY);
     }
 
     private static Block registerLegacyStair(BlockItemId id, Block base) {
-        return register(
-            id.block(),
-            p -> new StairBlock(base.defaultBlockState(), p),
-            BlockBehaviour.Properties.ofLegacyCopy(base).isViewBlocking(NEAR_PLANE_INTERSECTS_OUTLINE)
-        );
+        return register(id.block(), p -> new StairBlock(base.defaultBlockState(), p), BlockBehaviour.Properties.ofLegacyCopy(base));
     }
 
     private static Block registerStair(BlockItemId id, Block base) {
-        return register(
-            id, p -> new StairBlock(base.defaultBlockState(), p), BlockBehaviour.Properties.ofFullCopy(base).isViewBlocking(NEAR_PLANE_INTERSECTS_OUTLINE)
-        );
+        return register(id, p -> new StairBlock(base.defaultBlockState(), p), BlockBehaviour.Properties.ofFullCopy(base));
     }
 
     private static Block registerSlab(BlockItemId id, Block base) {
-        return register(id, SlabBlock::new, BlockBehaviour.Properties.ofLegacyCopy(base).isViewBlocking(NEAR_PLANE_INTERSECTS_OUTLINE));
+        return register(id, SlabBlock::new, BlockBehaviour.Properties.ofLegacyCopy(base));
     }
 
     private static Block registerSlab(BlockItemId id, Block base, float destroyTime) {
-        return register(id, SlabBlock::new, BlockBehaviour.Properties.ofLegacyCopy(base).isViewBlocking(NEAR_PLANE_INTERSECTS_OUTLINE).destroyTime(destroyTime));
+        return register(id, SlabBlock::new, BlockBehaviour.Properties.ofLegacyCopy(base).destroyTime(destroyTime));
     }
 
     private static Block registerSlab(BlockItemId id, Block base, float destroyTime, float explosionResistance) {
-        return register(
-            id,
-            SlabBlock::new,
-            BlockBehaviour.Properties.ofLegacyCopy(base).isViewBlocking(NEAR_PLANE_INTERSECTS_OUTLINE).strength(destroyTime, explosionResistance)
-        );
+        return register(id, SlabBlock::new, BlockBehaviour.Properties.ofLegacyCopy(base).strength(destroyTime, explosionResistance));
     }
 
     private static Block registerWall(BlockItemId id, Block base) {
