--- a/net/minecraft/world/level/block/BrewingStandBlock.java
+++ b/net/minecraft/world/level/block/BrewingStandBlock.java
@@ -77,8 +_,7 @@
             return InteractionResult.SUCCESS;
         } else {
             BlockEntity blockentity = p_50931_.getBlockEntity(p_50932_);
-            if (blockentity instanceof BrewingStandBlockEntity) {
-                p_50933_.openMenu((BrewingStandBlockEntity)blockentity);
+            if (blockentity instanceof BrewingStandBlockEntity && p_50933_.openMenu((BrewingStandBlockEntity)blockentity).isPresent()) { // Paper - Fix InventoryOpenEvent cancellation
                 p_50933_.awardStat(Stats.INTERACT_WITH_BREWINGSTAND);
             }
 
