--- a/net/minecraft/world/level/block/BuddingAmethystBlock.java
+++ b/net/minecraft/world/level/block/BuddingAmethystBlock.java
@@ -8,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluids;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class BuddingAmethystBlock extends AmethystBlock {
     public static final MapCodec<BuddingAmethystBlock> CODEC = simpleCodec(BuddingAmethystBlock::new);
@@ -44,7 +_,7 @@
                 BlockState blockstate1 = block.defaultBlockState()
                     .setValue(AmethystClusterBlock.FACING, direction)
                     .setValue(AmethystClusterBlock.WATERLOGGED, Boolean.valueOf(blockstate.getFluidState().getType() == Fluids.WATER));
-                p_220899_.setBlockAndUpdate(blockpos, blockstate1);
+                CraftEventFactory.handleBlockSpreadEvent(p_220899_, p_220900_, blockpos, blockstate1); // CraftBukkit
             }
         }
     }
