--- a/net/minecraft/world/level/block/BuddingAmethystBlock.java
+++ b/net/minecraft/world/level/block/BuddingAmethystBlock.java
@@ -44,7 +_,7 @@
                 BlockState targetState = nextStage.defaultBlockState()
                     .setValue(AmethystClusterBlock.FACING, growDirection)
                     .setValue(AmethystClusterBlock.WATERLOGGED, relativeState.getFluidState().is(Fluids.WATER));
-                level.setBlockAndUpdate(growPos, targetState);
+                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, growPos, targetState); // CraftBukkit
             }
         }
     }
