--- a/net/minecraft/world/level/block/BushBlock.java
+++ b/net/minecraft/world/level/block/BushBlock.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.pathfinder.PathComputationType;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public abstract class BushBlock extends Block {
     protected BushBlock(BlockBehaviour.Properties p_51021_) {
@@ -20,20 +_,28 @@
     protected abstract MapCodec<? extends BushBlock> codec();
 
     protected boolean mayPlaceOn(BlockState p_51042_, BlockGetter p_51043_, BlockPos p_51044_) {
-        return p_51042_.is(BlockTags.DIRT) || p_51042_.is(Blocks.FARMLAND);
+        return p_51042_.is(BlockTags.DIRT) || p_51042_.getBlock() instanceof net.minecraft.world.level.block.FarmBlock;
     }
 
     @Override
     protected BlockState updateShape(BlockState p_51032_, Direction p_51033_, BlockState p_51034_, LevelAccessor p_51035_, BlockPos p_51036_, BlockPos p_51037_) {
-        return !p_51032_.canSurvive(p_51035_, p_51036_)
-            ? Blocks.AIR.defaultBlockState()
-            : super.updateShape(p_51032_, p_51033_, p_51034_, p_51035_, p_51036_, p_51037_);
+        // CraftBukkit start
+        if (!p_51032_.canSurvive(p_51035_, p_51036_)) {
+            if (!(p_51035_ instanceof net.minecraft.server.level.ServerLevel && ((net.minecraft.server.level.ServerLevel) p_51035_).hasPhysicsEvent) || !org.bukkit.craftbukkit.event.CraftEventFactory.callBlockPhysicsEvent(p_51035_, p_51036_).isCancelled()) { // Paper
+                return Blocks.AIR.defaultBlockState();
+            }
+        }
+        return super.updateShape(p_51032_, p_51033_, p_51034_, p_51035_, p_51036_, p_51037_);
+        // CraftBukkit end
     }
 
     @Override
     protected boolean canSurvive(BlockState p_51028_, LevelReader p_51029_, BlockPos p_51030_) {
         BlockPos blockpos = p_51030_.below();
-        return this.mayPlaceOn(p_51029_.getBlockState(blockpos), p_51029_, blockpos);
+        BlockState belowBlockState = p_51029_.getBlockState(blockpos);
+        net.neoforged.neoforge.common.util.TriState soilDecision = belowBlockState.canSustainPlant(p_51029_, blockpos, Direction.UP, p_51028_);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
+        return this.mayPlaceOn(belowBlockState, p_51029_, blockpos);
     }
 
     @Override
