--- a/net/minecraft/world/level/block/BushBlock.java
+++ b/net/minecraft/world/level/block/BushBlock.java
@@ -20,7 +_,7 @@
         this(properties, 13);
     }
 
-    public BushBlock(BlockBehaviour.Properties properties, int shapeHeight) {
+    protected BushBlock(BlockBehaviour.Properties properties, int shapeHeight) {
         super(properties);
         this.shapeHeight = shapeHeight;
         this.shape = Block.column(16.0, 0.0, shapeHeight);
