--- a/net/minecraft/world/level/block/CampfireBlock.java
+++ b/net/minecraft/world/level/block/CampfireBlock.java
@@ -9,7 +_,6 @@
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.BlockTags;
-import net.minecraft.tags.ItemTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
@@ -88,23 +_,14 @@
             }
         }
 
-        if (itemStack.is(ItemTags.DOWSES_CAMPFIRES) && state.getValue(LIT)) {
-            if (!level.isClientSide()) {
-                level.levelEvent(null, 1009, pos, 0);
-            }
-
-            dowse(player, level, pos, state);
-            level.setBlockAndUpdate(pos, state.setValue(LIT, false));
-            return InteractionResult.SUCCESS;
-        } else {
-            return InteractionResult.TRY_WITH_EMPTY_HAND;
-        }
+        return InteractionResult.TRY_WITH_EMPTY_HAND;
     }
 
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier, boolean isPrecise) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (state.getValue(LIT) && entity instanceof LivingEntity) {
-            entity.hurt(level.damageSources().campfire(), this.fireDamage);
+            entity.hurt(level.damageSources().campfire().eventBlockDamager(level, pos), this.fireDamage); // CraftBukkit
         }
 
         super.entityInside(state, level, pos, entity, effectApplier, isPrecise);
@@ -118,7 +_,7 @@
         return this.defaultBlockState()
             .setValue(WATERLOGGED, replacedWater)
             .setValue(SIGNAL_FIRE, this.isSmokeSource(level.getBlockState(pos.below())))
-            .setValue(LIT, !replacedWater)
+            .setValue(LIT, level.getMinecraftWorld().purpurConfig.campFireLitWhenPlaced && !replacedWater) // Purpur - Campfire option for lit when placed
             .setValue(FACING, context.getHorizontalDirection());
     }
 
@@ -199,7 +_,7 @@
                 dowse(null, level, pos, state);
             }
 
-            level.setBlockAndUpdate(pos, state.setValue(WATERLOGGED, true).setValue(LIT, false));
+            level.setBlock(pos, state.setValue(WATERLOGGED, true).setValue(LIT, false), 3);
             level.scheduleTick(pos, fluidState.getType(), fluidState.getType().getTickDelay(level));
             return true;
         } else {
@@ -215,6 +_,11 @@
             && projectile.mayInteract(serverLevel, pos)
             && !state.getValue(LIT)
             && !state.getValue(WATERLOGGED)) {
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, pos, projectile).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
             level.setBlock(pos, state.setValue(BlockStateProperties.LIT, true), 11);
         }
     }
@@ -253,7 +_,7 @@
                 return true;
             }
 
-            boolean smokeBlocked = Shapes.joinIsNotEmpty(SHAPE_VIRTUAL_POST, blockState.getCollisionShape(level, pos, CollisionContext.empty()), BooleanOp.AND);
+            boolean smokeBlocked = Shapes.joinIsNotEmpty(SHAPE_VIRTUAL_POST, blockState.getCollisionShape(level, posToCheck, CollisionContext.empty()), BooleanOp.AND); // FORGE: Fix MC-201374
             if (smokeBlocked) {
                 BlockState belowState = level.getBlockState(posToCheck.below());
                 return isLitCampfire(belowState);
