--- a/net/minecraft/world/level/block/CampfireBlock.java
+++ b/net/minecraft/world/level/block/CampfireBlock.java
@@ -108,7 +_,7 @@
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier, boolean isPrecise) {
         if (state.getValue(LIT) && entity instanceof LivingEntity) {
-            entity.hurt(level.damageSources().campfire(), this.fireDamage);
+            entity.hurt(level.damageSources().campfire().directBlock(level, pos), this.fireDamage); // CraftBukkit
         }
 
         super.entityInside(state, level, pos, entity, effectApplier, isPrecise);
@@ -219,6 +_,11 @@
             && projectile.mayInteract(serverLevel, pos)
             && !state.getValue(LIT)
             && !state.getValue(WATERLOGGED)) {
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, pos, projectile).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
             level.setBlock(pos, state.setValue(BlockStateProperties.LIT, true), 11);
         }
     }
@@ -257,7 +_,7 @@
                 return true;
             }
 
-            boolean smokeBlocked = Shapes.joinIsNotEmpty(SHAPE_VIRTUAL_POST, blockState.getCollisionShape(level, pos, CollisionContext.empty()), BooleanOp.AND);
+            boolean smokeBlocked = Shapes.joinIsNotEmpty(SHAPE_VIRTUAL_POST, blockState.getCollisionShape(level, posToCheck, CollisionContext.empty()), BooleanOp.AND); // FORGE: Fix MC-201374
             if (smokeBlocked) {
                 BlockState belowState = level.getBlockState(posToCheck.below());
                 return isLitCampfire(belowState);
