--- a/net/minecraft/world/level/block/CampfireBlock.java
+++ b/net/minecraft/world/level/block/CampfireBlock.java
@@ -3,7 +_,6 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.Optional;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -48,6 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class CampfireBlock extends BaseEntityBlock implements SimpleWaterloggedBlock {
     public static final MapCodec<CampfireBlock> CODEC = RecordCodecBuilder.mapCodec(
@@ -109,8 +_,9 @@
 
     @Override
     protected void entityInside(BlockState p_51269_, Level p_51270_, BlockPos p_51271_, Entity p_51272_) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(p_51272_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_51270_, p_51271_)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (p_51269_.getValue(LIT) && p_51272_ instanceof LivingEntity) {
-            p_51272_.hurt(p_51270_.damageSources().campfire(), (float)this.fireDamage);
+            p_51272_.hurt(p_51270_.damageSources().campfire().directBlock(p_51270_, p_51271_), (float)this.fireDamage); // CraftBukkit
         }
 
         super.entityInside(p_51269_, p_51270_, p_51271_, p_51272_);
@@ -241,6 +_,11 @@
             && p_51247_.mayInteract(p_51244_, blockpos)
             && !p_51245_.getValue(LIT)
             && !p_51245_.getValue(WATERLOGGED)) {
+            // CraftBukkit start
+            if (CraftEventFactory.callBlockIgniteEvent(p_51244_, blockpos, p_51247_).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
             p_51244_.setBlock(blockpos, p_51245_.setValue(BlockStateProperties.LIT, Boolean.valueOf(true)), 11);
         }
     }
@@ -279,7 +_,7 @@
                 return true;
             }
 
-            boolean flag = Shapes.joinIsNotEmpty(VIRTUAL_FENCE_POST, blockstate.getCollisionShape(p_51249_, p_51250_, CollisionContext.empty()), BooleanOp.AND);
+            boolean flag = Shapes.joinIsNotEmpty(VIRTUAL_FENCE_POST, blockstate.getCollisionShape(p_51249_, blockpos, CollisionContext.empty()), BooleanOp.AND); // FORGE: Fix MC-201374
             if (flag) {
                 BlockState blockstate1 = p_51249_.getBlockState(blockpos.below());
                 return isLitCampfire(blockstate1);
