--- a/net/minecraft/world/level/block/CarvedPumpkinBlock.java
+++ b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
@@ -26,6 +_,7 @@
 import net.minecraft.world.level.block.state.pattern.BlockPatternBuilder;
 import net.minecraft.world.level.block.state.predicate.BlockStatePredicate;
 import net.minecraft.world.level.block.state.properties.EnumProperty;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.jspecify.annotations.Nullable;
 
 public class CarvedPumpkinBlock extends HorizontalDirectionalBlock {
@@ -104,16 +_,22 @@
                 .getAge();
     }
 
-    private static void spawnGolemInWorld(Level level, BlockPattern.BlockPatternMatch match, Entity golem, BlockPos spawnPos) {
-        clearPatternBlocks(level, match);
+    // CraftBukkit start
+    private static boolean spawnGolemInWorld(Level level, BlockPattern.BlockPatternMatch match, Entity golem, BlockPos spawnPos) {
+        // clearPatternBlocks(level, match); // CraftBukkit - moved down
         golem.snapTo(spawnPos.getX() + 0.5, spawnPos.getY() + 0.05, spawnPos.getZ() + 0.5, 0.0F, 0.0F);
-        level.addFreshEntity(golem);
+        if (!level.addFreshEntity(golem, (golem.getType() == EntityTypes.SNOW_GOLEM) ? CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN : ((golem.getType() == EntityTypes.IRON_GOLEM) ? CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM : CreatureSpawnEvent.SpawnReason.BUILD_COPPERGOLEM))) {
+            return false;
+        }
+        clearPatternBlocks(level, match); // CraftBukkit - from above
+        // CraftBukkit end
 
         for (ServerPlayer player : level.getEntitiesOfClass(ServerPlayer.class, golem.getBoundingBox().inflate(5.0))) {
             CriteriaTriggers.SUMMONED_ENTITY.trigger(player, golem);
         }
 
         updatePatternBlocks(level, match);
+        return true; // CraftBukkit
     }
 
     public static void clearPatternBlocks(Level level, BlockPattern.BlockPatternMatch match) {
