--- a/net/minecraft/world/level/block/CarvedPumpkinBlock.java
+++ b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
@@ -104,11 +_,24 @@
                 .getAge();
     }
 
+    // CraftBukkit start
     private static void spawnGolemInWorld(Level level, BlockPattern.BlockPatternMatch match, Entity golem, BlockPos spawnPos) {
-        clearPatternBlocks(level, match);
+        // clearPatternBlocks(level, patternMatch); // Paper - moved down
         golem.snapTo(spawnPos.getX() + 0.5, spawnPos.getY() + 0.05, spawnPos.getZ() + 0.5, 0.0F, 0.0F);
-        level.addFreshEntity(golem);
-
+        // Paper start
+        org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason;
+        if (golem.getType() == net.minecraft.world.entity.EntityTypes.SNOW_GOLEM) {
+            spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN;
+        } else if (golem.getType() == net.minecraft.world.entity.EntityTypes.COPPER_GOLEM) {
+            spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_COPPERGOLEM;
+        } else {
+            spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM;
+        }
+        if (!level.addFreshEntity(golem, spawnReason)) {
+            return;
+        }
+        // Paper end
+        clearPatternBlocks(level, match); // Paper - from above
         for (ServerPlayer player : level.getEntitiesOfClass(ServerPlayer.class, golem.getBoundingBox().inflate(5.0))) {
             CriteriaTriggers.SUMMONED_ENTITY.trigger(player, golem);
         }
