--- a/net/minecraft/world/level/block/CarvedPumpkinBlock.java
+++ b/net/minecraft/world/level/block/CarvedPumpkinBlock.java
@@ -67,7 +_,7 @@
         if (snowGolemMatch != null) {
             SnowGolem snowGolem = EntityTypes.SNOW_GOLEM.create(level, EntitySpawnReason.TRIGGERED);
             if (snowGolem != null) {
-                spawnGolemInWorld(level, snowGolemMatch, snowGolem, snowGolemMatch.getBlock(0, 2, 0).getPos());
+                spawnGolemInWorld(level, snowGolemMatch, snowGolem, snowGolemMatch.getBlock(0, 2, 0).getPos(), this.placer); // Purpur - Summoner API
                 return;
             }
         }
@@ -77,7 +_,7 @@
             IronGolem ironGolem = EntityTypes.IRON_GOLEM.create(level, EntitySpawnReason.TRIGGERED);
             if (ironGolem != null) {
                 ironGolem.setPlayerCreated(true);
-                spawnGolemInWorld(level, ironGolemMatch, ironGolem, ironGolemMatch.getBlock(1, 2, 0).getPos());
+                spawnGolemInWorld(level, ironGolemMatch, ironGolem, ironGolemMatch.getBlock(1, 2, 0).getPos(), this.placer); // Purpur - Summoner API
                 return;
             }
         }
@@ -86,7 +_,7 @@
         if (copperGolemMatch != null) {
             CopperGolem copperGolem = EntityTypes.COPPER_GOLEM.create(level, EntitySpawnReason.TRIGGERED);
             if (copperGolem != null) {
-                spawnGolemInWorld(level, copperGolemMatch, copperGolem, copperGolemMatch.getBlock(0, 0, 0).getPos());
+                spawnGolemInWorld(level, copperGolemMatch, copperGolem, copperGolemMatch.getBlock(0, 0, 0).getPos(), this.placer); // Purpur - Summoner API
                 this.replaceCopperBlockWithChest(level, copperGolemMatch);
                 copperGolem.spawn(this.getWeatherStateFromPattern(copperGolemMatch));
             }
@@ -104,11 +_,39 @@
                 .getAge();
     }
 
+    @io.papermc.paper.annotation.DoNotUse
     private static void spawnGolemInWorld(Level level, BlockPattern.BlockPatternMatch match, Entity golem, BlockPos spawnPos) {
-        clearPatternBlocks(level, match);
+        // Purpur start - Summoner API
+        spawnGolemInWorld(level, match, golem, spawnPos, null);
+    }
+    private static void spawnGolemInWorld(final Level level, final BlockPattern.BlockPatternMatch match, final Entity golem, final BlockPos spawnPos, final net.minecraft.world.entity.LivingEntity placer) {
+        java.util.UUID summoner = placer == null ? null : placer.getUUID();
+        switch (golem) {
+            case SnowGolem snowGolem -> snowGolem.setSummoner(summoner);
+            case IronGolem ironGolem -> ironGolem.setSummoner(summoner);
+            case CopperGolem copperGolem -> copperGolem.setSummoner(summoner);
+            default -> throw new IllegalStateException("Unexpected value: " + golem);
+        }
+        // Purpur end - Summoner API
+        // clearPatternBlocks(level, patternMatch); // Paper - moved down
         golem.snapTo(spawnPos.getX() + 0.5, spawnPos.getY() + 0.05, spawnPos.getZ() + 0.5, 0.0F, 0.0F);
-        level.addFreshEntity(golem);
-
+        // Paper start
+        if (!new io.papermc.paper.event.entity.EntityConstructEvent(golem.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.getMatchingBlocks(level, match)).callEvent()) {
+            return;
+        }
+        org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason;
+        if (golem.getType() == net.minecraft.world.entity.EntityTypes.SNOW_GOLEM) {
+            spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_SNOWMAN;
+        } else if (golem.getType() == net.minecraft.world.entity.EntityTypes.COPPER_GOLEM) {
+            spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_COPPERGOLEM;
+        } else {
+            spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.BUILD_IRONGOLEM;
+        }
+        if (!level.addFreshEntity(golem, spawnReason)) {
+            return;
+        }
+        // Paper end
+        clearPatternBlocks(level, match); // Paper - from above
         for (ServerPlayer player : level.getEntitiesOfClass(ServerPlayer.class, golem.getBoundingBox().inflate(5.0))) {
             CriteriaTriggers.SUMMONED_ENTITY.trigger(player, golem);
         }
