--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -32,9 +_,9 @@
 
     protected static boolean shouldHandlePrecipitation(Level level, Biome.Precipitation precipitation) {
         if (precipitation == Biome.Precipitation.RAIN) {
-            return level.getRandom().nextFloat() < 0.05F;
+            return level.getRandom().nextFloat() < level.purpurConfig.cauldronRainChance;
         } else {
-            return precipitation == Biome.Precipitation.SNOW ? level.getRandom().nextFloat() < 0.1F : false;
+            return precipitation == Biome.Precipitation.SNOW ? level.getRandom().nextFloat() <level.purpurConfig.cauldronPowderSnowChance : false;
         }
     }
 
@@ -42,10 +_,18 @@
     public void handlePrecipitation(BlockState state, Level level, BlockPos pos, Biome.Precipitation precipitation) {
         if (shouldHandlePrecipitation(level, precipitation)) {
             if (precipitation == Biome.Precipitation.RAIN) {
-                level.setBlockAndUpdate(pos, Blocks.WATER_CAULDRON.defaultBlockState());
+                // Paper start - Call CauldronLevelChangeEvent
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleCauldronLevelChangeEvent(level, pos, Blocks.WATER_CAULDRON.defaultBlockState(), null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL)) { // avoid duplicate game event
+                    return;
+                }
+                // Paper end - Call CauldronLevelChangeEvent
                 level.gameEvent(null, GameEvent.BLOCK_CHANGE, pos);
             } else if (precipitation == Biome.Precipitation.SNOW) {
-                level.setBlockAndUpdate(pos, Blocks.POWDER_SNOW_CAULDRON.defaultBlockState());
+                // Paper start - Call CauldronLevelChangeEvent
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleCauldronLevelChangeEvent(level, pos, Blocks.POWDER_SNOW_CAULDRON.defaultBlockState(), null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL)) { // avoid duplicate game event
+                    return;
+                }
+                // Paper end - Call CauldronLevelChangeEvent
                 level.gameEvent(null, GameEvent.BLOCK_CHANGE, pos);
             }
         }
@@ -58,16 +_,25 @@
 
     @Override
     protected void receiveStalactiteDrip(BlockState state, Level level, BlockPos pos, Fluid fluid) {
+        if (fluid.getFluidType().handleCauldronDrip(fluid, level, pos)) return;
         if (fluid == Fluids.WATER) {
             BlockState newState = Blocks.WATER_CAULDRON.defaultBlockState();
-            level.setBlockAndUpdate(pos, newState);
+            // Paper start - Call CauldronLevelChangeEvent; don't send level event or game event if cancelled
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleCauldronLevelChangeEvent(level, pos, newState, null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL)) { // CraftBukkit
+                return;
+            }
+            // Paper end - Call CauldronLevelChangeEvent
             level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(newState));
-            level.levelEvent(1047, pos, 0);
+            level.levelEvent(LevelEvent.SOUND_DRIP_WATER_INTO_CAULDRON, pos, 0);
         } else if (fluid == Fluids.LAVA) {
             BlockState newState = Blocks.LAVA_CAULDRON.defaultBlockState();
-            level.setBlockAndUpdate(pos, newState);
+            // Paper start - Call CauldronLevelChangeEvent; don't send level event or game event if cancelled
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleCauldronLevelChangeEvent(level, pos, newState, null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL)) { // CraftBukkit
+                return;
+            }
+            // Paper end - Call CauldronLevelChangeEvent
             level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(newState));
-            level.levelEvent(1046, pos, 0);
+            level.levelEvent(LevelEvent.SOUND_DRIP_LAVA_INTO_CAULDRON, pos, 0);
         }
     }
 }
