--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -10,6 +_,7 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
+import org.bukkit.event.block.CauldronLevelChangeEvent;
 
 public class CauldronBlock extends AbstractCauldronBlock {
     public static final MapCodec<CauldronBlock> CODEC = simpleCodec(CauldronBlock::new);
@@ -58,16 +_,25 @@
 
     @Override
     protected void receiveStalactiteDrip(BlockState state, Level level, BlockPos pos, Fluid fluid) {
+        if (fluid.getFluidType().handleCauldronDrip(fluid, level, pos)) return;
         if (fluid == Fluids.WATER) {
             BlockState newState = Blocks.WATER_CAULDRON.defaultBlockState();
-            level.setBlockAndUpdate(pos, newState);
+            // Paper start - Call CauldronLevelChangeEvent; don't send level event or game event if cancelled
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleCauldronLevelChangeEvent(level, pos, newState, null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL)) { // CraftBukkit
+                return;
+            }
+            // Paper end - Call CauldronLevelChangeEvent
             level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(newState));
-            level.levelEvent(1047, pos, 0);
+            level.levelEvent(LevelEvent.SOUND_DRIP_WATER_INTO_CAULDRON, pos, 0);
         } else if (fluid == Fluids.LAVA) {
             BlockState newState = Blocks.LAVA_CAULDRON.defaultBlockState();
-            level.setBlockAndUpdate(pos, newState);
+            // Paper start - Call CauldronLevelChangeEvent; don't send level event or game event if cancelled
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleCauldronLevelChangeEvent(level, pos, newState, null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL)) { // CraftBukkit
+                return;
+            }
+            // Paper end - Call CauldronLevelChangeEvent
             level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(newState));
-            level.levelEvent(1046, pos, 0);
+            level.levelEvent(LevelEvent.SOUND_DRIP_LAVA_INTO_CAULDRON, pos, 0);
         }
     }
 }
