--- a/net/minecraft/world/level/block/CaveVines.java
+++ b/net/minecraft/world/level/block/CaveVines.java
@@ -16,6 +_,16 @@
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+// CraftBukkit start
+import java.util.LinkedList;
+import java.util.List;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.ItemStack;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerHarvestBlockEvent;
+// CraftBukkit end
+
 public interface CaveVines {
     VoxelShape SHAPE = Block.column(14.0, 0.0, 16.0);
     BooleanProperty BERRIES = BlockStateProperties.BERRIES;
@@ -23,15 +_,29 @@
     static InteractionResult use(Entity sourceEntity, BlockState state, Level level, BlockPos pos) {
         if (state.getValue(BERRIES)) {
             if (level instanceof ServerLevel serverLevel) {
-                Block.dropFromBlockInteractLootTable(
-                    serverLevel,
-                    BuiltInLootTables.HARVEST_CAVE_VINE,
-                    state,
-                    level.getBlockEntity(pos),
-                    null,
-                    sourceEntity,
-                    (serverlvl, itemStack) -> Block.popResource(serverlvl, pos, itemStack)
-                );
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(sourceEntity, pos, state.setValue(CaveVines.BERRIES, false))) {
+                    return InteractionResult.SUCCESS;
+                }
+
+                if (sourceEntity instanceof Player) {
+                    List<ItemStack> dropped = new LinkedList<>();
+                    Block.dropFromBlockInteractLootTable(serverLevel, BuiltInLootTables.HARVEST_CAVE_VINE, state, level.getBlockEntity(pos), null, sourceEntity, (serverlvl, itemstack) -> {
+                        dropped.add(itemstack);
+                    });
+                    PlayerHarvestBlockEvent event = CraftEventFactory.callPlayerHarvestBlockEvent(level, pos, (Player) sourceEntity, net.minecraft.world.InteractionHand.MAIN_HAND, dropped);
+                    if (event.isCancelled()) {
+                        return InteractionResult.SUCCESS; // We need to return a success either way, because making it PASS or FAIL will result in a bug where cancelling while harvesting w/ block in hand places block
+                    }
+                    for (org.bukkit.inventory.ItemStack itemStack : event.getItemsHarvested()) {
+                        Block.popResource(level, pos, CraftItemStack.asNMSCopy(itemStack));
+                    }
+                } else {
+                    Block.dropFromBlockInteractLootTable(serverLevel, BuiltInLootTables.HARVEST_CAVE_VINE, state, level.getBlockEntity(pos), null, sourceEntity, (serverlvl, itemstack) -> {
+                        Block.popResource(serverlvl, pos, itemstack);
+                    });
+                }
+                // CraftBukkit end
                 float pitch = Mth.randomBetween(serverLevel.getRandom(), 0.8F, 1.2F);
                 serverLevel.playSound(null, pos, SoundEvents.CAVE_VINES_PICK_BERRIES, SoundSource.BLOCKS, 1.0F, pitch);
                 BlockState newState = state.setValue(BERRIES, false);
