--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -5,6 +_,7 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public interface ChangeOverTimeBlock<T extends Enum<T>> {
     int SCAN_DISTANCE = 4;
@@ -16,7 +_,9 @@
     default void changeOverTime(BlockState p_309154_, ServerLevel p_309115_, BlockPos p_308999_, RandomSource p_308868_) {
         float f = 0.05688889F;
         if (p_308868_.nextFloat() < 0.05688889F) {
-            this.getNextState(p_309154_, p_309115_, p_308999_, p_308868_).ifPresent(p_153039_ -> p_309115_.setBlockAndUpdate(p_308999_, p_153039_));
+            this.getNextState(p_309154_, p_309115_, p_308999_, p_308868_).ifPresent(p_153039_ ->
+                CraftEventFactory.handleBlockFormEvent(p_309115_, p_308999_, p_153039_) // CraftBukkit
+            );
         }
     }
 
