--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -3,10 +_,8 @@
 import java.util.Optional;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.util.Continuation;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.block.state.BlockState;
-import org.apache.commons.lang3.mutable.MutableInt;
 
 public interface ChangeOverTimeBlock<T extends Enum<T>> {
     int SCAN_DISTANCE = 4;
@@ -18,7 +_,7 @@
     default void changeOverTime(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         float eachBlockOncePerDayChance = 0.05688889F;
         if (random.nextFloat() < 0.05688889F) {
-            this.getNextState(state, level, pos, random).ifPresent(weatheredState -> level.setBlockAndUpdate(pos, weatheredState));
+            this.getNextState(state, level, pos, random).ifPresent(weatheredState -> org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, weatheredState, Block.UPDATE_ALL)); // CraftBukkit
         }
     }
 
@@ -26,37 +_,34 @@
 
     default Optional<BlockState> getNextState(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         int ownAge = this.getAge().ordinal();
-        MutableInt sameAgeCount = new MutableInt(0);
-        MutableInt olderCount = new MutableInt(0);
-        boolean aborted = level.findBlocksInManhattan(pos, 4)
-            .filterPos(blockPos -> !blockPos.equals(pos))
-            .filterState(this::isSameAgeingType)
-            .forEachUntil((blockPos, var5x) -> {
-                if (level.getBlockState(blockPos).getBlock() instanceof ChangeOverTimeBlock<?> neighborBlock) {
-                    int foundAge = neighborBlock.getAge().ordinal();
+        int sameAgeCount = 0;
+        int olderCount = 0;
+
+        for (BlockPos blockPos : BlockPos.withinManhattan(pos, 4, 4, 4)) {
+            int manhattanDistance = blockPos.distManhattan(pos);
+            if (manhattanDistance > 4) {
+                break;
+            }
+
+            if (!blockPos.equals(pos) && level.getBlockState(blockPos).getBlock() instanceof ChangeOverTimeBlock<?> neighborBlock) {
+                Enum<?> neighborAge = neighborBlock.getAge();
+                if (this.getAge().getClass() == neighborAge.getClass()) {
+                    int foundAge = neighborAge.ordinal();
                     if (foundAge < ownAge) {
-                        return Continuation.ABORT;
+                        return Optional.empty();
                     }
 
                     if (foundAge > ownAge) {
-                        olderCount.increment();
+                        olderCount++;
                     } else {
-                        sameAgeCount.increment();
+                        sameAgeCount++;
                     }
                 }
-
-                return Continuation.CONTINUE;
-            });
-        if (aborted) {
-            return Optional.empty();
+            }
         }
 
-        float chance = (float)(olderCount.intValue() + 1) / (olderCount.intValue() + sameAgeCount.intValue() + 1);
-        float actualChance = chance * chance * this.getChanceModifier();
+        float chance = (float)(olderCount + 1) / (olderCount + sameAgeCount + 1);
+        float actualChance = level.purpurConfig.disableOxidationProximityPenalty ? this.getChanceModifier() : chance * chance * this.getChanceModifier(); // Purpur - option to disable the copper oxidation proximity penalty
         return random.nextFloat() < actualChance ? this.getNext(state) : Optional.empty();
-    }
-
-    private boolean isSameAgeingType(BlockState state) {
-        return state.getBlock() instanceof ChangeOverTimeBlock<?> neighborBlock && this.getAge().getClass() == neighborBlock.getAge().getClass();
     }
 }
