--- a/net/minecraft/world/level/block/ChiseledBookShelfBlock.java
+++ b/net/minecraft/world/level/block/ChiseledBookShelfBlock.java
@@ -108,7 +_,7 @@
         }
     }
 
-    public OptionalInt getHitSlot(BlockHitResult p_316156_, BlockState p_316148_) {
+    private OptionalInt getHitSlot(BlockHitResult p_316156_, BlockState p_316148_) {
         return getRelativeHitCoordinatesForBlockFace(p_316156_, p_316148_.getValue(HorizontalDirectionalBlock.FACING)).map(p_316073_ -> {
             int i = p_316073_.y >= 0.5F ? 0 : 1;
             int j = getSection(p_316073_.x);
@@ -137,7 +_,7 @@
         }
     }
 
-    public static int getSection(float p_261599_) {
+    private static int getSection(float p_261599_) {
         float f = 0.0625F;
         float f1 = 0.375F;
         if (p_261599_ < 0.375F) {
