--- a/net/minecraft/world/level/block/CommandBlock.java
+++ b/net/minecraft/world/level/block/CommandBlock.java
@@ -64,6 +_,17 @@
             if (p_51839_.getBlockEntity(p_51840_) instanceof CommandBlockEntity commandblockentity) {
                 boolean flag = p_51839_.hasNeighborSignal(p_51840_);
                 boolean flag1 = commandblockentity.isPowered();
+
+                // CraftBukkit start
+                org.bukkit.block.Block bukkitBlock = p_51839_.getWorld().getBlockAt(p_51840_.getX(), p_51840_.getY(), p_51840_.getZ());
+                int old = flag1 ? 15 : 0;
+                int current = flag ? 15 : 0;
+
+                org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(bukkitBlock, old, current);
+                p_51839_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+                flag = eventRedstone.getNewCurrent() > 0;
+                // CraftBukkit end
+
                 commandblockentity.setPowered(flag);
                 if (!flag1 && !commandblockentity.isAutomatic() && commandblockentity.getMode() != CommandBlockEntity.Mode.SEQUENCE) {
                     if (flag) {
@@ -118,7 +_,7 @@
     @Override
     protected InteractionResult useWithoutItem(BlockState p_51825_, Level p_51826_, BlockPos p_51827_, Player p_51828_, BlockHitResult p_51830_) {
         BlockEntity blockentity = p_51826_.getBlockEntity(p_51827_);
-        if (blockentity instanceof CommandBlockEntity && p_51828_.canUseGameMasterBlocks()) {
+        if (blockentity instanceof CommandBlockEntity && (p_51828_.canUseGameMasterBlocks() || (p_51828_.isCreative() && p_51828_.getBukkitEntity().hasPermission("minecraft.commandblock")))) { // Paper - command block permission
             p_51828_.openCommandBlock((CommandBlockEntity)blockentity);
             return InteractionResult.sidedSuccess(p_51826_.isClientSide);
         } else {
