--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.level.block;
 
+import com.google.common.util.concurrent.AtomicDouble;
 import com.mojang.serialization.MapCodec;
 import it.unimi.dsi.fastutil.objects.Object2FloatMap;
 import it.unimi.dsi.fastutil.objects.Object2FloatOpenHashMap;
@@ -41,6 +_,9 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftBlockInventoryHolder;
+import org.bukkit.craftbukkit.util.DummyGeneratorAccess;
 
 public class ComposterBlock extends Block implements WorldlyContainerHolder {
     public static final MapCodec<ComposterBlock> CODEC = simpleCodec(ComposterBlock::new);
@@ -48,6 +_,8 @@
     public static final int MIN_LEVEL = 0;
     public static final int MAX_LEVEL = 7;
     public static final IntegerProperty LEVEL = BlockStateProperties.LEVEL_COMPOSTER;
+    /** @deprecated Neo: Use the {@link net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps#COMPOSTABLES compostable} data map instead */
+    @Deprecated
     public static final Object2FloatMap<ItemLike> COMPOSTABLES = new Object2FloatOpenHashMap<>();
     private static final int AABB_SIDE_THICKNESS = 2;
     private static final VoxelShape OUTER_SHAPE = Shapes.block();
@@ -227,6 +_,15 @@
         if (p_51978_.getValue(LEVEL) == 7) {
             p_51979_.scheduleTick(p_51980_, p_51978_.getBlock(), 20);
         }
+        // Neo: Invalidate composter capabilities when a composter is added
+        if (!p_51981_.is(this)) p_51979_.invalidateCapabilities(p_51980_);
+    }
+
+    @Override
+    protected void onRemove(BlockState p_60515_, Level p_60516_, BlockPos p_60517_, BlockState p_60518_, boolean p_60519_) {
+        super.onRemove(p_60515_, p_60516_, p_60517_, p_60518_, p_60519_);
+        // Neo: Invalidate composter capabilities when a composter is removed
+        if (!p_60515_.is(p_60518_.getBlock())) p_60516_.invalidateCapabilities(p_60517_);
     }
 
     @Override
@@ -234,7 +_,7 @@
         ItemStack p_316332_, BlockState p_316118_, Level p_316624_, BlockPos p_316660_, Player p_316715_, InteractionHand p_316472_, BlockHitResult p_316606_
     ) {
         int i = p_316118_.getValue(LEVEL);
-        if (i < 8 && COMPOSTABLES.containsKey(p_316332_.getItem())) {
+        if (i < 8 && getValue(p_316332_) > 0) {
             if (i < 7 && !p_316624_.isClientSide) {
                 BlockState blockstate = addItem(p_316715_, p_316118_, p_316624_, p_316660_, p_316332_);
                 p_316624_.levelEvent(1500, p_316660_, p_316118_ != blockstate ? 1 : 0);
@@ -261,8 +_,15 @@
 
     public static BlockState insertItem(Entity p_270919_, BlockState p_270087_, ServerLevel p_270284_, ItemStack p_270253_, BlockPos p_270678_) {
         int i = p_270087_.getValue(LEVEL);
-        if (i < 7 && COMPOSTABLES.containsKey(p_270253_.getItem())) {
-            BlockState blockstate = addItem(p_270919_, p_270087_, p_270284_, p_270678_, p_270253_);
+        if (i < 7 && getValue(p_270253_) > 0) {
+            // CraftBukkit start
+            double rand = p_270284_.getRandom().nextDouble();
+            BlockState blockstate = addItem(p_270919_, p_270087_, DummyGeneratorAccess.INSTANCE, p_270678_, p_270253_, rand);
+            if (p_270087_ == blockstate || !CraftEventFactory.callEntityChangeBlockEvent(p_270919_, p_270678_, blockstate)) {
+                return p_270087_;
+            }
+            blockstate = addItem(p_270919_, p_270087_, p_270284_, p_270678_, p_270253_, rand);
+            // CraftBukkit end
             p_270253_.shrink(1);
             return blockstate;
         } else {
@@ -271,6 +_,14 @@
     }
 
     public static BlockState extractProduce(Entity p_270467_, BlockState p_51999_, Level p_52000_, BlockPos p_52001_) {
+        // CraftBukkit start
+        if (p_270467_ != null && !(p_270467_ instanceof Player)) {
+            BlockState blockstate1 = empty(p_270467_, p_51999_, DummyGeneratorAccess.INSTANCE, p_52001_);
+            if (!CraftEventFactory.callEntityChangeBlockEvent(p_270467_, p_52001_, blockstate1)) {
+                return p_51999_;
+            }
+        }
+        // CraftBukkit end
         if (!p_52000_.isClientSide) {
             Vec3 vec3 = Vec3.atLowerCornerWithOffset(p_52001_, 0.5, 1.01, 0.5).offsetRandom(p_52000_.random, 0.7F);
             ItemEntity itementity = new ItemEntity(p_52000_, vec3.x(), vec3.y(), vec3.z(), new ItemStack(Items.BONE_MEAL));
@@ -290,10 +_,29 @@
         return blockstate;
     }
 
+    static AtomicDouble bukkitRand = new AtomicDouble(-114514);
     static BlockState addItem(@Nullable Entity p_270464_, BlockState p_270603_, LevelAccessor p_270151_, BlockPos p_270547_, ItemStack p_270354_) {
+        // CraftBukkit end
         int i = p_270603_.getValue(LEVEL);
-        float f = COMPOSTABLES.getFloat(p_270354_.getItem());
-        if ((i != 0 || !(f > 0.0F)) && !(p_270151_.getRandom().nextDouble() < (double)f)) {
+        float f = getValue(p_270354_);
+        if (bukkitRand.get() == -114514) {
+            bukkitRand.set(p_270151_.getRandom().nextDouble());
+        }
+        // Paper start - Add CompostItemEvent and EntityCompostItemEvent
+        boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && bukkitRand.getAndSet(-114514) >= (double) f);
+        final io.papermc.paper.event.block.CompostItemEvent event;
+        if (p_270464_ == null) {
+            event = new io.papermc.paper.event.block.CompostItemEvent(org.bukkit.craftbukkit.block.CraftBlock.at(p_270151_, p_270547_), p_270354_.getBukkitStack(), willRaiseLevel);
+        } else {
+            event = new io.papermc.paper.event.entity.EntityCompostItemEvent(p_270464_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_270151_, p_270547_), p_270354_.getBukkitStack(), willRaiseLevel);
+        }
+        if (!event.callEvent()) { // check for cancellation of entity event (non entity event can't be cancelled cause of hoppers)
+            return null;
+        }
+        willRaiseLevel = event.willRaiseLevel();
+
+        if (!willRaiseLevel) {
+            // Paper end - Add CompostItemEvent and EntityCompostItemEvent
             return p_270603_;
         } else {
             int j = i + 1;
@@ -306,6 +_,13 @@
 
             return blockstate;
         }
+
+    }
+
+    static BlockState addItem(@Nullable Entity p_270464_, BlockState p_270603_, LevelAccessor p_270151_, BlockPos p_270547_, ItemStack p_270354_, double rand) {
+        bukkitRand.set(rand);
+        // CraftBukkit start
+        return addItem(p_270464_, p_270603_, p_270151_, p_270547_, p_270354_);
     }
 
     @Override
@@ -342,14 +_,22 @@
         if (i == 8) {
             return new ComposterBlock.OutputContainer(p_51956_, p_51957_, p_51958_, new ItemStack(Items.BONE_MEAL));
         } else {
-            return (WorldlyContainer)(i < 7 ? new ComposterBlock.InputContainer(p_51956_, p_51957_, p_51958_) : new ComposterBlock.EmptyContainer());
+            return (WorldlyContainer)(i < 7 ? new ComposterBlock.InputContainer(p_51956_, p_51957_, p_51958_) : new ComposterBlock.EmptyContainer(p_51957_, p_51958_)); // CraftBukkit - add parameters
         }
     }
 
     public static class EmptyContainer extends SimpleContainer implements WorldlyContainer {
-        public EmptyContainer() {
+        // CraftBukkit start
+        private final LevelAccessor level;
+        private final BlockPos pos;
+
+        public EmptyContainer(LevelAccessor p_52043_, BlockPos p_52044_) {
             super(0);
+            this.level = p_52043_;
+            this.pos = p_52044_;
+            this.bukkitOwner = new CraftBlockInventoryHolder(p_52043_, p_52044_, this);
         }
+        // CraftBukkit end
 
         @Override
         public int[] getSlotsForFace(Direction p_52012_) {
@@ -375,6 +_,7 @@
 
         public InputContainer(BlockState p_52022_, LevelAccessor p_52023_, BlockPos p_52024_) {
             super(1);
+            this.bukkitOwner = new CraftBlockInventoryHolder(p_52023_, p_52024_, this); // CraftBukkit
             this.state = p_52022_;
             this.level = p_52023_;
             this.pos = p_52024_;
@@ -392,7 +_,7 @@
 
         @Override
         public boolean canPlaceItemThroughFace(int p_52028_, ItemStack p_52029_, @Nullable Direction p_52030_) {
-            return !this.changed && p_52030_ == Direction.UP && ComposterBlock.COMPOSTABLES.containsKey(p_52029_.getItem());
+            return !this.changed && p_52030_ == Direction.UP && getValue(p_52029_) > 0f;
         }
 
         @Override
@@ -406,6 +_,11 @@
             if (!itemstack.isEmpty()) {
                 this.changed = true;
                 BlockState blockstate = ComposterBlock.addItem(null, this.state, this.level, this.pos, itemstack);
+                // Paper start - Add CompostItemEvent and EntityCompostItemEvent
+                if (blockstate == null) {
+                    return;
+                }
+                // Paper end - Add CompostItemEvent and EntityCompostItemEvent
                 this.level.levelEvent(1500, this.pos, blockstate != this.state ? 1 : 0);
                 this.removeItemNoUpdate(0);
             }
@@ -420,6 +_,7 @@
 
         public OutputContainer(BlockState p_52042_, LevelAccessor p_52043_, BlockPos p_52044_, ItemStack p_52045_) {
             super(p_52045_);
+            this.bukkitOwner = new CraftBlockInventoryHolder(p_52043_, p_52044_, this); // CraftBukkit
             this.state = p_52042_;
             this.level = p_52043_;
             this.pos = p_52044_;
@@ -447,8 +_,21 @@
 
         @Override
         public void setChanged() {
-            ComposterBlock.empty(null, this.state, this.level, this.pos);
-            this.changed = true;
+            // CraftBukkit start - allow putting items back (eg cancelled InventoryMoveItemEvent)
+            if (this.isEmpty()) {
+                ComposterBlock.empty(null, this.state, this.level, this.pos);
+                this.changed = true;
+            } else {
+                this.level.setBlock(this.pos, this.state, 3);
+                this.changed = false;
+            }
+            // CraftBukkit end
         }
+    }
+
+    public static float getValue(ItemStack item) {
+        var value = item.getItemHolder().getData(net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps.COMPOSTABLES);
+        if (value != null) return value.chance();
+        return -1f;
     }
 }
