--- a/net/minecraft/world/level/block/ConcretePowderBlock.java
+++ b/net/minecraft/world/level/block/ConcretePowderBlock.java
@@ -23,33 +_,54 @@
 
     @Override
     public void onLand(Level level, BlockPos pos, BlockState state, BlockState replacedBlock, FallingBlockEntity entity) {
-        if (shouldSolidify(level, pos, replacedBlock)) {
-            level.setBlockAndUpdate(pos, this.concrete.defaultBlockState());
+        if (shouldSolidify(level, pos, state, replacedBlock.getFluidState())) { // Forge: Use block of falling entity instead of block at replaced position, and check if shouldSolidify with the FluidState of the replaced block
+            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, this.concrete.defaultBlockState(), 3); // CraftBukkit
         }
     }
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
-        BlockGetter level = context.getLevel();
+        Level level = context.getLevel(); // Paper
         BlockPos pos = context.getClickedPos();
         BlockState replacedBlock = level.getBlockState(pos);
-        return shouldSolidify(level, pos, replacedBlock) ? this.concrete.defaultBlockState() : super.getStateForPlacement(context);
+        // CraftBukkit start
+        if (!ConcretePowderBlock.shouldSolidify(level, pos, replacedBlock)) {
+            return super.getStateForPlacement(context);
+        }
+
+        // TODO: An event factory call for methods like this
+        org.bukkit.craftbukkit.block.CraftBlockState snapshot = org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(level, pos);
+        snapshot.setBlock(this.concrete.defaultBlockState());
+
+        org.bukkit.event.block.BlockFormEvent event = new org.bukkit.event.block.BlockFormEvent(snapshot.getBlock(), snapshot);
+        event.callEvent();
+
+        if (!event.isCancelled()) {
+            return snapshot.getHandle();
+        }
+
+        return super.getStateForPlacement(context);
+        // CraftBukkit end
+    }
+
+    private static boolean shouldSolidify(BlockGetter level, BlockPos pos, BlockState replacedBlock, net.minecraft.world.level.material.FluidState fluidState) {
+        return replacedBlock.canBeHydrated(level, pos, fluidState, pos) || touchesLiquid(level, pos, replacedBlock);
     }
 
     private static boolean shouldSolidify(BlockGetter level, BlockPos pos, BlockState replacedBlock) {
-        return canSolidify(replacedBlock) || touchesLiquid(level, pos);
+        return shouldSolidify(level, pos, replacedBlock, level.getFluidState(pos));
     }
 
-    private static boolean touchesLiquid(BlockGetter level, BlockPos pos) {
+    private static boolean touchesLiquid(BlockGetter level, BlockPos pos, BlockState state) {
         boolean touchesLiquid = false;
         BlockPos.MutableBlockPos testPos = pos.mutable();
 
         for (Direction direction : Direction.values()) {
             BlockState blockState = level.getBlockState(testPos);
-            if (direction != Direction.DOWN || canSolidify(blockState)) {
+            if (direction != Direction.DOWN || state.canBeHydrated(level, pos, blockState.getFluidState(), testPos)) {
                 testPos.setWithOffset(pos, direction);
                 blockState = level.getBlockState(testPos);
-                if (canSolidify(blockState) && !blockState.isFaceSturdy(level, pos, direction.getOpposite())) {
+                if (state.canBeHydrated(level, pos, blockState.getFluidState(), testPos) && !blockState.isFaceSturdy(level, pos, direction.getOpposite())) {
                     touchesLiquid = true;
                     break;
                 }
@@ -74,9 +_,25 @@
         BlockState neighbourState,
         RandomSource random
     ) {
-        return touchesLiquid(level, pos)
-            ? this.concrete.defaultBlockState()
-            : super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
+        // CraftBukkit start
+        if (ConcretePowderBlock.touchesLiquid(level, pos, state)) { // NeoForge
+            // Suppress during worldgen
+            if (!(level instanceof Level world1)) {
+                return this.concrete.defaultBlockState();
+            }
+            org.bukkit.craftbukkit.block.CraftBlockState snapshot = org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(world1, pos);
+            snapshot.setBlock(this.concrete.defaultBlockState());
+
+            org.bukkit.event.block.BlockFormEvent event = new org.bukkit.event.block.BlockFormEvent(snapshot.getBlock(), snapshot);
+            world1.getCraftServer().getPluginManager().callEvent(event);
+
+            if (!event.isCancelled()) {
+                return snapshot.getHandle();
+            }
+        }
+
+        return super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
+        // CraftBukkit end
     }
 
     @Override
