--- a/net/minecraft/world/level/block/CopperBulbBlock.java
+++ b/net/minecraft/world/level/block/CopperBulbBlock.java
@@ -40,13 +_,18 @@
     public void checkAndFlip(BlockState state, ServerLevel level, BlockPos pos) {
         boolean signal = level.hasNeighborSignal(pos);
         if (signal != state.getValue(POWERED)) {
+            // Paper start - Call BlockRedstoneEvent
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBinaryRedstoneChange(level, pos, signal)) {
+                return;
+            }
+            // Paper end - Call BlockRedstoneEvent
             BlockState newState = state;
             if (!state.getValue(POWERED)) {
                 newState = newState.cycle(LIT);
                 level.playSound(null, pos, newState.getValue(LIT) ? SoundEvents.COPPER_BULB_TURN_ON : SoundEvents.COPPER_BULB_TURN_OFF, SoundSource.BLOCKS);
             }
 
-            level.setBlockAndUpdate(pos, newState.setValue(POWERED, signal));
+            level.setBlock(pos, newState.setValue(POWERED, signal), 3);
         }
     }
 
