--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -2,13 +_,11 @@
 
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.tags.FluidTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.item.context.BlockPlaceContext;
 import net.minecraft.world.level.BlockGetter;
@@ -16,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class CoralBlock extends Block {
     public static final MapCodec<Block> DEAD_CORAL_FIELD = BuiltInRegistries.BLOCK.byNameCodec().fieldOf("dead");
@@ -37,6 +_,11 @@
     @Override
     protected void tick(BlockState p_221020_, ServerLevel p_221021_, BlockPos p_221022_, RandomSource p_221023_) {
         if (!this.scanForWater(p_221021_, p_221022_)) {
+            // CraftBukkit start
+            if (CraftEventFactory.callBlockFadeEvent(p_221021_, p_221022_, this.deadBlock.defaultBlockState()).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
             p_221021_.setBlock(p_221022_, this.deadBlock.defaultBlockState(), 2);
         }
     }
@@ -51,9 +_,10 @@
     }
 
     protected boolean scanForWater(BlockGetter p_52135_, BlockPos p_52136_) {
+        BlockState state = p_52135_.getBlockState(p_52136_);
         for (Direction direction : Direction.values()) {
             FluidState fluidstate = p_52135_.getFluidState(p_52136_.relative(direction));
-            if (fluidstate.is(FluidTags.WATER)) {
+            if (state.canBeHydrated(p_52135_, p_52136_, fluidstate, p_52136_.relative(direction))) {
                 return true;
             }
         }
