--- a/net/minecraft/world/level/block/CraftingTableBlock.java
+++ b/net/minecraft/world/level/block/CraftingTableBlock.java
@@ -33,11 +_,10 @@
     protected InteractionResult useWithoutItem(BlockState p_52233_, Level p_52234_, BlockPos p_52235_, Player p_52236_, BlockHitResult p_52238_) {
         if (p_52234_.isClientSide) {
             return InteractionResult.SUCCESS;
-        } else {
-            p_52236_.openMenu(p_52233_.getMenuProvider(p_52234_, p_52235_));
+        } else if (p_52236_.openMenu(p_52233_.getMenuProvider(p_52234_, p_52235_)).isPresent()) { // Paper - Fix InventoryOpenEvent cancellation
             p_52236_.awardStat(Stats.INTERACT_WITH_CRAFTING_TABLE);
-            return InteractionResult.CONSUME;
         }
+        return InteractionResult.CONSUME; // Paper - Fix InventoryOpenEvent cancellation
     }
 
     @Override
