--- a/net/minecraft/world/level/block/DispenserBlock.java
+++ b/net/minecraft/world/level/block/DispenserBlock.java
@@ -54,6 +_,7 @@
     private static final DefaultDispenseItemBehavior DEFAULT_BEHAVIOR = new DefaultDispenseItemBehavior();
     public static final Map<Item, DispenseItemBehavior> DISPENSER_REGISTRY = new IdentityHashMap<>();
     private static final int TRIGGER_DURATION = 4;
+    public static boolean eventFired = false; // CraftBukkit
 
     @Override
     public MapCodec<? extends DispenserBlock> codec() {
@@ -97,11 +_,18 @@
                 ItemStack itemStack = blockEntity.getItem(slot);
                 DispenseItemBehavior behavior = this.getDispenseMethod(level, itemStack);
                 if (behavior != DispenseItemBehavior.NOOP) {
+                    eventFired = false; // CraftBukkit - reset event status
                     blockEntity.setItem(slot, behavior.dispense(source, itemStack));
                 }
             }
         }
     }
+
+    // Paper start - Fix NPE with equippable and items without behavior
+    public static DispenseItemBehavior getDispenseBehavior(BlockSource pointer, ItemStack stack) {
+        return ((DispenserBlock) pointer.state().getBlock()).getDispenseMethod(pointer.level(), stack);
+    }
+    // Paper end - Fix NPE with equippable and items without behavior
 
     protected DispenseItemBehavior getDispenseMethod(Level level, ItemStack itemStack) {
         if (!itemStack.isItemEnabled(level.enabledFeatures())) {
