--- a/net/minecraft/world/level/block/DispenserBlock.java
+++ b/net/minecraft/world/level/block/DispenserBlock.java
@@ -49,6 +_,7 @@
         new Object2ObjectOpenHashMap<>(), p_340795_ -> p_340795_.defaultReturnValue(DEFAULT_BEHAVIOR)
     );
     private static final int TRIGGER_DURATION = 4;
+    public static boolean eventFired = false; // CraftBukkit
 
     @Override
     public MapCodec<? extends DispenserBlock> codec() {
@@ -95,12 +_,16 @@
             BlockSource blocksource = new BlockSource(p_52665_, p_52666_, p_302451_, dispenserblockentity);
             int i = dispenserblockentity.getRandomSlot(p_52665_.random);
             if (i < 0) {
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFailedDispenseEvent(p_52665_, p_52666_)) { // Paper - Add BlockFailedDispenseEvent
                 p_52665_.levelEvent(1001, p_52666_, 0);
                 p_52665_.gameEvent(GameEvent.BLOCK_ACTIVATE, p_52666_, GameEvent.Context.of(dispenserblockentity.getBlockState()));
+                } // Paper - Add BlockFailedDispenseEvent
             } else {
                 ItemStack itemstack = dispenserblockentity.getItem(i);
                 DispenseItemBehavior dispenseitembehavior = this.getDispenseMethod(p_52665_, itemstack);
                 if (dispenseitembehavior != DispenseItemBehavior.NOOP) {
+                    if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockPreDispenseEvent(p_52665_, p_52666_, itemstack, i)) return; // Paper - Add BlockPreDispenseEvent
+                    eventFired = false; // CraftBukkit - reset event status
                     dispenserblockentity.setItem(i, dispenseitembehavior.dispense(blocksource, itemstack));
                 }
             }
