--- a/net/minecraft/world/level/block/DispenserBlock.java
+++ b/net/minecraft/world/level/block/DispenserBlock.java
@@ -87,7 +_,7 @@
         }
     }
 
-    public void dispenseFrom(ServerLevel p_52665_, BlockState p_302451_, BlockPos p_52666_) {
+    protected void dispenseFrom(ServerLevel p_52665_, BlockState p_302451_, BlockPos p_52666_) {
         DispenserBlockEntity dispenserblockentity = p_52665_.getBlockEntity(p_52666_, BlockEntityType.DISPENSER).orElse(null);
         if (dispenserblockentity == null) {
             LOGGER.warn("Ignoring dispensing attempt for Dispenser without matching block entity at {}", p_52666_);
