--- a/net/minecraft/world/level/block/DoublePlantBlock.java
+++ b/net/minecraft/world/level/block/DoublePlantBlock.java
@@ -73,6 +_,7 @@
         }
 
         BlockState belowState = level.getBlockState(pos.below());
+        if (state.getBlock() != this) return super.canSurvive(state, level, pos); // Neo: This function is called during world gen and placement, before this block is set, so if we are not 'here' then assume it's the pre-check.
         return belowState.is(this) && belowState.getValue(HALF) == DoubleBlockHalf.LOWER;
     }
 
@@ -100,11 +_,16 @@
     }
 
     @Override
-    public void playerDestroy(Level level, Player player, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, ItemStack destroyedWith) {
-        super.playerDestroy(level, player, pos, Blocks.AIR.defaultBlockState(), blockEntity, destroyedWith);
+    public void playerDestroy(Level level, Player player, BlockPos pos, BlockState state, @Nullable BlockEntity blockEntity, ItemStack destroyedWith, boolean includeDrops, boolean dropExp) { // Paper - fix drops not preventing stats/food exhaustion
+        super.playerDestroy(level, player, pos, Blocks.AIR.defaultBlockState(), blockEntity, destroyedWith, includeDrops, dropExp); // Paper - fix drops not preventing stats/food exhaustion;
     }
 
     protected static void preventDropFromBottomPart(Level level, BlockPos pos, BlockState state, Player player) {
+        // CraftBukkit start
+        if (((net.minecraft.server.level.ServerLevel)level).hasPhysicsEvent && org.bukkit.craftbukkit.event.CraftEventFactory.callBlockPhysicsEvent(level, pos).isCancelled()) { // Paper
+            return;
+        }
+        // CraftBukkit end
         DoubleBlockHalf part = state.getValue(HALF);
         if (part == DoubleBlockHalf.UPPER) {
             BlockPos bottomPos = pos.below();
