--- a/net/minecraft/world/level/block/DragonEggBlock.java
+++ b/net/minecraft/world/level/block/DragonEggBlock.java
@@ -1,7 +_,8 @@
 package net.minecraft.world.level.block;
 
 import net.minecraft.core.BlockPos;
-import net.minecraft.util.BlockUtil;
+import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.util.Mth;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.player.Player;
@@ -16,8 +_,6 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 
 public class DragonEggBlock extends FallingBlock {
-    public static final int HORIZONTAL_TELEPORT_RADIUS = 16;
-    public static final int VERTICAL_TELEPORT_RADIUS = 8;
     private static final VoxelShape SHAPE = Block.column(14.0, 0.0, 16.0);
 
     public DragonEggBlock(BlockBehaviour.Properties properties) {
@@ -41,24 +_,45 @@
     }
 
     private void teleport(BlockState state, Level level, BlockPos pos) {
-        if (!level.isClientSide()) {
-            WorldBorder worldBorder = level.getWorldBorder();
-            RandomSource random = level.getRandom();
-
-            for (int i = 0; i < 1000; i++) {
-                BlockPos testPos = pos.offset(
-                    random.nextInt(16) - random.nextInt(16), random.nextInt(8) - random.nextInt(8), random.nextInt(16) - random.nextInt(16)
-                );
-                if (level.getBlockState(testPos).isAir()
-                    && !level.getBlockState(testPos.below()).isAir()
-                    && worldBorder.isWithinBounds(testPos)
-                    && level.isInsideBuildHeight(testPos)) {
-                    int packedDiff = BlockUtil.packDifferenceInPosition(pos, testPos, 16, 8, 16);
-                    level.levelEvent(2015, pos, packedDiff);
+        if (!level.purpurConfig.dragonEggTeleport) return; // Purpur - Option to disable dragon egg teleporting
+        WorldBorder worldBorder = level.getWorldBorder();
+        RandomSource random = level.getRandom();
+
+        for (int i = 0; i < 1000; i++) {
+            BlockPos testPos = pos.offset(
+                random.nextInt(16) - random.nextInt(16), random.nextInt(8) - random.nextInt(8), random.nextInt(16) - random.nextInt(16)
+            );
+            if (level.getBlockState(testPos).isAir()
+                && !level.getBlockState(testPos.below()).isAir()
+                && worldBorder.isWithinBounds(testPos)
+                && level.isInsideBuildHeight(testPos)) {
+                // CraftBukkit start
+                org.bukkit.block.Block from = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos);
+                org.bukkit.block.Block to = org.bukkit.craftbukkit.block.CraftBlock.at(level, testPos);
+                org.bukkit.event.block.BlockFromToEvent event = new org.bukkit.event.block.BlockFromToEvent(from, to);
+                if (!event.callEvent()) {
+                    return;
+                }
+
+                testPos = new BlockPos(event.getToBlock().getX(), event.getToBlock().getY(), event.getToBlock().getZ());
+                // CraftBukkit end
+                if (level.isClientSide()) {
+                    for (int j = 0; j < 128; j++) {
+                        double d = random.nextDouble();
+                        float xa = (random.nextFloat() - 0.5F) * 0.2F;
+                        float ya = (random.nextFloat() - 0.5F) * 0.2F;
+                        float za = (random.nextFloat() - 0.5F) * 0.2F;
+                        double x = Mth.lerp(d, testPos.getX(), pos.getX()) + (random.nextDouble() - 0.5) + 0.5;
+                        double y = Mth.lerp(d, testPos.getY(), pos.getY()) + random.nextDouble() - 0.5;
+                        double z = Mth.lerp(d, testPos.getZ(), pos.getZ()) + (random.nextDouble() - 0.5) + 0.5;
+                        level.addParticle(ParticleTypes.PORTAL, x, y, z, xa, ya, za);
+                    }
+                } else {
                     level.setBlock(testPos, state, 2);
                     level.removeBlock(pos, false);
-                    return;
                 }
+
+                return;
             }
         }
     }
