--- a/net/minecraft/world/level/block/DropperBlock.java
+++ b/net/minecraft/world/level/block/DropperBlock.java
@@ -18,12 +_,14 @@
 import net.minecraft.world.level.block.entity.HopperBlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.slf4j.Logger;
 
 public class DropperBlock extends DispenserBlock {
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final MapCodec<DropperBlock> CODEC = simpleCodec(DropperBlock::new);
-    private static final DispenseItemBehavior DISPENSE_BEHAVIOUR = new DefaultDispenseItemBehavior();
+    private static final DispenseItemBehavior DISPENSE_BEHAVIOUR = new DefaultDispenseItemBehavior(true); // CraftBukkit
 
     @Override
     public MapCodec<DropperBlock> codec() {
@@ -53,18 +_,37 @@
             BlockSource blocksource = new BlockSource(p_52944_, p_52945_, p_302455_, dispenserblockentity);
             int i = dispenserblockentity.getRandomSlot(p_52944_.random);
             if (i < 0) {
-                p_52944_.levelEvent(1001, p_52945_, 0);
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFailedDispenseEvent(p_52944_, p_52945_)) // Paper - Add BlockFailedDispenseEvent
+                    p_52944_.levelEvent(1001, p_52945_, 0);
             } else {
                 ItemStack itemstack = dispenserblockentity.getItem(i);
-                if (!itemstack.isEmpty()) {
+                if (!itemstack.isEmpty() && net.neoforged.neoforge.items.VanillaInventoryCodeHooks.dropperInsertHook(p_52944_, p_52945_, dispenserblockentity, i, itemstack)) {
                     Direction direction = p_52944_.getBlockState(p_52945_).getValue(FACING);
                     Container container = HopperBlockEntity.getContainerAt(p_52944_, p_52945_.relative(direction));
                     ItemStack itemstack1;
                     if (container == null) {
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockPreDispenseEvent(p_52944_, p_52945_, itemstack, i)) return; // Paper - Add BlockPreDispenseEvent
                         itemstack1 = DISPENSE_BEHAVIOUR.dispense(blocksource, itemstack);
                     } else {
-                        itemstack1 = HopperBlockEntity.addItem(dispenserblockentity, container, itemstack.copyWithCount(1), direction.getOpposite());
-                        if (itemstack1.isEmpty()) {
+                        // CraftBukkit start - Fire event when pushing items into other inventories
+                        CraftItemStack oitemstack = CraftItemStack.asCraftMirror(itemstack.copyWithCount(1));
+
+                        org.bukkit.inventory.Inventory destinationInventory;
+                        // Have to special case large chests as they work oddly
+                        if (container instanceof net.minecraft.world.CompoundContainer) {
+                            destinationInventory = new CraftInventoryDoubleChest((net.minecraft.world.CompoundContainer) container);
+                        } else {
+                            destinationInventory = container.getOwner().getInventory();
+                        }
+
+                        org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(dispenserblockentity.getOwner().getInventory(), oitemstack.clone(), destinationInventory, true);
+                        p_52944_.getCraftServer().getPluginManager().callEvent(event);
+                        if (event.isCancelled()) {
+                            return;
+                        }
+                        itemstack1 = HopperBlockEntity.addItem(dispenserblockentity, container, CraftItemStack.asNMSCopy(event.getItem()), direction.getOpposite());
+                        if (event.getItem().equals(oitemstack) && itemstack1.isEmpty()) {
+                        // CraftBukkit end
                             itemstack1 = itemstack.copy();
                             itemstack1.shrink(1);
                         } else {
