--- a/net/minecraft/world/level/block/DropperBlock.java
+++ b/net/minecraft/world/level/block/DropperBlock.java
@@ -7,7 +_,6 @@
 import net.minecraft.core.dispenser.DefaultDispenseItemBehavior;
 import net.minecraft.core.dispenser.DispenseItemBehavior;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.world.Container;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.BlockEntity;
@@ -38,7 +_,7 @@
     }
 
     @Override
-    protected void dispenseFrom(ServerLevel level, BlockState state, BlockPos pos) {
+    public void dispenseFrom(ServerLevel level, BlockState state, BlockPos pos) {
         DispenserBlockEntity blockEntity = level.getBlockEntity(pos, BlockEntityTypes.DROPPER).orElse(null);
         if (blockEntity == null) {
             LOGGER.warn("Ignoring dispensing attempt for Dropper without matching block entity at {}", pos);
@@ -46,22 +_,51 @@
             BlockSource source = new BlockSource(level, pos, state, blockEntity);
             int slot = blockEntity.getRandomSlot(level.getRandom());
             if (slot < 0) {
-                level.levelEvent(1001, pos, 0);
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFailedDispenseEvent(level, pos)) // Paper - Add BlockFailedDispenseEvent
+                    level.levelEvent(1001, pos, 0);
             } else {
                 ItemStack itemStack = blockEntity.getItem(slot);
                 if (!itemStack.isEmpty()) {
                     Direction direction = level.getBlockState(pos).getValue(FACING);
-                    Container into = HopperBlockEntity.getContainerAt(level, pos.relative(direction));
+                    var into = HopperBlockEntity.getContainerOrHandlerAt(level, pos.relative(direction), direction.getOpposite());
                     ItemStack remaining;
-                    if (into == null) {
+                    if (into.isEmpty()) {
+                        if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockPreDispenseEvent(level, pos, itemStack, slot)) return; // Paper - Add BlockPreDispenseEvent
                         remaining = DISPENSE_BEHAVIOUR.dispense(source, itemStack);
                     } else {
-                        remaining = HopperBlockEntity.addItem(blockEntity, into, itemStack.copyWithCount(1), direction.getOpposite());
-                        if (remaining.isEmpty()) {
-                            remaining = itemStack.copy();
-                            remaining.shrink(1);
+                        if (into.container() != null) {
+                            // CraftBukkit start - Fire event when pushing items into other inventories
+                            org.bukkit.craftbukkit.inventory.CraftItemStack oitemstack = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack.copyWithCount(1));
+
+                            org.bukkit.inventory.Inventory destinationInventory;
+                            // Have to special case large chests as they work oddly
+                            if (into.container() instanceof net.minecraft.world.CompoundContainer compoundContainer) {
+                                destinationInventory = new org.bukkit.craftbukkit.inventory.CraftInventoryDoubleChest(compoundContainer);
+                            } else {
+                                destinationInventory = into.container().getOwner().getInventory();
+                            }
+
+                            org.bukkit.event.inventory.InventoryMoveItemEvent event = new org.bukkit.event.inventory.InventoryMoveItemEvent(blockEntity.getOwner().getInventory(), oitemstack, destinationInventory, true);
+                            if (!event.callEvent()) {
+                                return;
+                            }
+                            remaining = HopperBlockEntity.addItem(blockEntity, into.container(), org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getItem()), direction.getOpposite());
+                            if (event.getItem().equals(oitemstack) && remaining.isEmpty()) {
+                                // CraftBukkit end
+                                remaining = itemStack.copy();
+                                remaining.shrink(1);
+                            } else {
+                                // CraftBukkit end
+                                remaining = itemStack.copy();
+                            }
                         } else {
-                            remaining = itemStack.copy();
+                            remaining = net.neoforged.neoforge.transfer.item.ItemUtil.insertItemReturnRemaining(into.itemHandler(), itemStack.copyWithCount(1), false, null);
+                            if (remaining.isEmpty()) {
+                                remaining = itemStack.copy();
+                                remaining.shrink(1);
+                            } else {
+                                remaining = itemStack.copy();
+                            }
                         }
                     }
 
