--- a/net/minecraft/world/level/block/EndGatewayBlock.java
+++ b/net/minecraft/world/level/block/EndGatewayBlock.java
@@ -22,6 +_,7 @@
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.portal.TeleportTransition;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.player.PlayerTeleportEvent;
 import org.jspecify.annotations.Nullable;
 
 public class EndGatewayBlock extends BaseEntityBlock implements Portal {
@@ -79,10 +_,15 @@
 
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier, boolean isPrecise) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (entity.canUsePortal(false)
             && !level.isClientSide()
             && level.getBlockEntity(pos) instanceof TheEndGatewayBlockEntity endGatewayBlockEntity
             && !endGatewayBlockEntity.isCoolingDown()) {
+            // Paper start - call EntityPortalEnterEvent
+            org.bukkit.event.entity.EntityPortalEnterEvent event = new org.bukkit.event.entity.EntityPortalEnterEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level), org.bukkit.PortalType.END_GATEWAY); // Paper - add portal type
+            if (!event.callEvent()) return;
+            // Paper end - call EntityPortalEnterEvent
             entity.setAsInsidePortal(this, pos);
             TheEndGatewayBlockEntity.triggerCooldown(level, pos, state, endGatewayBlockEntity);
         }
@@ -96,7 +_,7 @@
                 return null;
             } else {
                 return entity instanceof ThrownEnderpearl
-                    ? new TeleportTransition(currentLevel, teleportPosition, Vec3.ZERO, 0.0F, 0.0F, Set.of(), TeleportTransition.PLACE_PORTAL_TICKET)
+                    ? new TeleportTransition(currentLevel, teleportPosition, Vec3.ZERO, 0.0F, 0.0F, Set.of(), TeleportTransition.PLACE_PORTAL_TICKET, PlayerTeleportEvent.TeleportCause.END_GATEWAY)
                     : new TeleportTransition(
                         currentLevel,
                         teleportPosition,
@@ -104,7 +_,8 @@
                         0.0F,
                         0.0F,
                         Relative.union(Relative.DELTA, Relative.ROTATION),
-                        TeleportTransition.PLACE_PORTAL_TICKET
+                        TeleportTransition.PLACE_PORTAL_TICKET,
+                        PlayerTeleportEvent.TeleportCause.END_GATEWAY
                     );
             }
         } else {
