--- a/net/minecraft/world/level/block/EndGatewayBlock.java
+++ b/net/minecraft/world/level/block/EndGatewayBlock.java
@@ -85,6 +_,7 @@
 
     @Override
     protected void entityInside(BlockState p_350647_, Level p_350785_, BlockPos p_350610_, Entity p_350849_) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(p_350849_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_350785_, p_350610_)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (p_350849_.canUsePortal(false)
             && !p_350785_.isClientSide
             && p_350785_.getBlockEntity(p_350610_) instanceof TheEndGatewayBlockEntity theendgatewayblockentity
