--- a/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/net/minecraft/world/level/block/EndPortalBlock.java
@@ -20,6 +_,7 @@
 import net.minecraft.world.level.block.entity.TheEndPortalBlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.levelgen.feature.EndPlatformFeature;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.portal.TeleportTransition;
@@ -29,6 +_,16 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.event.CraftPortalEvent;
+import org.bukkit.craftbukkit.util.CraftLocation;
+import org.bukkit.event.entity.EntityPortalEnterEvent;
+import org.bukkit.event.player.PlayerRespawnEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
+// CraftBukkit end
+
 public class EndPortalBlock extends BaseEntityBlock implements Portal {
     public static final MapCodec<EndPortalBlock> CODEC = simpleCodec(EndPortalBlock::new);
     private static final VoxelShape SHAPE = Block.column(16.0, 6.0, 12.0);
@@ -60,6 +_,10 @@
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier, boolean isPrecise) {
         if (entity.canUsePortal(false)) {
+            // CraftBukkit start - Entity in portal
+            org.bukkit.event.entity.EntityPortalEnterEvent event = new org.bukkit.event.entity.EntityPortalEnterEvent(entity.getBukkitEntity(), new org.bukkit.Location(level.getWorld(), pos.getX(), pos.getY(), pos.getZ()));
+            level.getCraftServer().getPluginManager().callEvent(event);
+            // CraftBukkit end
             if (!level.isClientSide() && level.dimension() == Level.END && entity instanceof ServerPlayer player && !player.seenCredits) {
                 player.showEndCredits();
             } else {
@@ -70,42 +_,51 @@
 
     @Override
     public @Nullable TeleportTransition getPortalDestination(ServerLevel currentLevel, Entity entity, BlockPos portalEntryPos) {
-        LevelData.RespawnData respawnData = currentLevel.getRespawnData();
-        ResourceKey<Level> currentDimension = currentLevel.dimension();
-        boolean fromEnd = currentDimension == Level.END;
+        // CraftBukkit start - SPIGOT-8111: send back to main overworld
+        LevelData.RespawnData respawnData = currentLevel.getServer().findRespawnDimension().getRespawnData();
+        // SPIGOT-6152: send back to main overworld in custom ends
+        ResourceKey<LevelStem> resourcekey = currentLevel.getTypeKey();
+        boolean fromEnd = resourcekey == LevelStem.END;
+        // CraftBukkit end
         ResourceKey<Level> newDimension = fromEnd ? respawnData.dimension() : Level.END;
         BlockPos spawnBlockPos = fromEnd ? respawnData.pos() : ServerLevel.END_SPAWN_POINT;
         ServerLevel newLevel = currentLevel.getServer().getLevel(newDimension);
         if (newLevel == null) {
-            return null;
-        }
-
-        Vec3 spawnPos = Vec3.atBottomCenterOf(spawnBlockPos);
-        float yRot;
-        float xRot;
-        Set<Relative> relatives;
-        if (!fromEnd) {
-            EndPlatformFeature.createEndPlatform(newLevel, BlockPos.containing(spawnPos).below(), true);
-            yRot = Direction.WEST.toYRot();
-            xRot = 0.0F;
-            relatives = Relative.union(Relative.DELTA, Set.of(Relative.X_ROT));
-            if (entity instanceof ServerPlayer) {
-                spawnPos = spawnPos.subtract(0.0, 1.0, 0.0);
-            }
+            return new TeleportTransition(PlayerTeleportEvent.TeleportCause.END_PORTAL); // CraftBukkit
         } else {
-            yRot = respawnData.yaw();
-            xRot = respawnData.pitch();
-            relatives = Relative.union(Relative.DELTA, Relative.ROTATION);
-            if (entity instanceof ServerPlayer serverPlayer) {
-                return serverPlayer.findRespawnPositionAndUseSpawnBlock(false, TeleportTransition.DO_NOTHING);
-            }
-
-            spawnPos = Vec3.atBottomCenterOf(entity.adjustSpawnLocation(newLevel, spawnBlockPos));
+            Vec3 spawnPos = spawnBlockPos.getBottomCenter();
+            float yRot;
+            float xRot;
+            Set<Relative> relatives;
+            if (!fromEnd) {
+                EndPlatformFeature.createEndPlatform(newLevel, BlockPos.containing(spawnPos).below(), true);
+                yRot = Direction.WEST.toYRot();
+                xRot = 0.0F;
+                relatives = Relative.union(Relative.DELTA, Set.of(Relative.X_ROT));
+                if (entity instanceof ServerPlayer) {
+                    spawnPos = spawnPos.subtract(0.0, 1.0, 0.0);
+                }
+            } else {
+                yRot = respawnData.yaw();
+                xRot = respawnData.pitch();
+                relatives = Relative.union(Relative.DELTA, Relative.ROTATION);
+                if (entity instanceof ServerPlayer serverPlayer) {
+                    return serverPlayer.findRespawnPositionAndUseSpawnBlock(false, TeleportTransition.DO_NOTHING, PlayerRespawnEvent.RespawnReason.END_PORTAL); // CraftBukkit
+                }
+
+                spawnPos = Vec3.atBottomCenterOf(entity.adjustSpawnLocation(newLevel, spawnBlockPos));
+            }
+
+            // CraftBukkit start
+            CraftPortalEvent event = entity.callPortalEvent(entity, CraftLocation.toBukkit(spawnPos, newLevel.getWorld(), yRot, xRot), PlayerTeleportEvent.TeleportCause.END_PORTAL, 0, 0);
+            if (event == null) {
+                return null;
+            }
+            Location to = event.getTo();
+
+            return new TeleportTransition(((CraftWorld) to.getWorld()).getHandle(), CraftLocation.toVec3D(to), Vec3.ZERO, to.getYaw(), to.getPitch(), relatives, TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET), PlayerTeleportEvent.TeleportCause.END_PORTAL);
+            // CraftBukkit end
         }
-
-        return new TeleportTransition(
-            newLevel, spawnPos, Vec3.ZERO, yRot, xRot, relatives, TeleportTransition.PLAY_PORTAL_SOUND.then(TeleportTransition.PLACE_PORTAL_TICKET)
-        );
     }
 
     @Override
