--- a/net/minecraft/world/level/block/EyeblossomBlock.java
+++ b/net/minecraft/world/level/block/EyeblossomBlock.java
@@ -80,25 +_,31 @@
         boolean shouldBeOpen = level.environmentAttributes().getValue(EnvironmentAttributes.EYEBLOSSOM_OPEN, pos).toBoolean(this.type.open);
         if (shouldBeOpen == this.type.open) {
             return false;
-        }
-
-        EyeblossomBlock.Type newType = this.type.transform();
-        level.setBlock(pos, newType.state(), 3);
-        level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(state));
-        newType.spawnTransformParticle(level, pos, random);
-        BlockPos.betweenClosed(pos.offset(-3, -2, -3), pos.offset(3, 2, 3)).forEach(nearby -> {
-            BlockState nearbyState = level.getBlockState(nearby);
-            if (nearbyState == state) {
-                double distance = Math.sqrt(pos.distSqr(nearby));
-                int delay = random.nextIntBetweenInclusive((int)(distance * 5.0), (int)(distance * 10.0));
-                level.scheduleTick(nearby, state.getBlock(), delay);
+        } else {
+            EyeblossomBlock.Type newType = this.type.transform();
+            // CraftBukkit start - BlockFormEvent
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, newType.state(), 3)) {
+                return false;
             }
-        });
-        return true;
+            // CraftBukkit end
+            level.setBlock(pos, newType.state(), 3);
+            level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(state));
+            newType.spawnTransformParticle(level, pos, random);
+            BlockPos.betweenClosed(pos.offset(-3, -2, -3), pos.offset(3, 2, 3)).forEach(nearby -> {
+                BlockState nearbyState = level.getBlockState(nearby);
+                if (nearbyState == state) {
+                    double distance = Math.sqrt(pos.distSqr(nearby));
+                    int delay = random.nextIntBetweenInclusive((int)(distance * 5.0), (int)(distance * 10.0));
+                    level.scheduleTick(nearby, state.getBlock(), delay);
+                }
+            });
+            return true;
+        }
     }
 
     @Override
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier, boolean isPrecise) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (!level.isClientSide()
             && level.getDifficulty() != Difficulty.PEACEFUL
             && entity instanceof Bee bee
