--- a/net/minecraft/world/level/block/FlowerBedBlock.java
+++ b/net/minecraft/world/level/block/FlowerBedBlock.java
@@ -25,7 +_,7 @@
     private final int shapeHeight;
     private final Function<BlockState, VoxelShape> shapes;
 
-    public FlowerBedBlock(BlockBehaviour.Properties properties, int shapeHeight) {
+    protected FlowerBedBlock(BlockBehaviour.Properties properties, int shapeHeight) {
         super(properties);
         this.shapeHeight = shapeHeight;
         this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(AMOUNT, 1));
