--- a/net/minecraft/world/level/block/FrogspawnBlock.java
+++ b/net/minecraft/world/level/block/FrogspawnBlock.java
@@ -113,6 +_,7 @@
                 int k = p_221223_.nextInt(1, 361);
                 tadpole.moveTo(d0, (double)p_221222_.getY() - 0.5, d1, (float)k, 0.0F);
                 tadpole.setPersistenceRequired();
+                tadpole.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EGG); // Paper - use correct spawn reason
                 p_221221_.addFreshEntity(tadpole);
             }
         }
