--- a/net/minecraft/world/level/block/FrogspawnBlock.java
+++ b/net/minecraft/world/level/block/FrogspawnBlock.java
@@ -81,6 +_,7 @@
 
     @Override
     protected void entityInside(BlockState p_221204_, Level p_221205_, BlockPos p_221206_, Entity p_221207_) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(p_221207_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_221205_, p_221206_)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (p_221207_.getType().equals(EntityType.FALLING_BLOCK)) {
             this.destroyBlock(p_221205_, p_221206_);
         }
@@ -113,6 +_,7 @@
                 int k = p_221223_.nextInt(1, 361);
                 tadpole.moveTo(d0, (double)p_221222_.getY() - 0.5, d1, (float)k, 0.0F);
                 tadpole.setPersistenceRequired();
+                tadpole.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EGG); // Paper - use correct spawn reason
                 p_221221_.addFreshEntity(tadpole);
             }
         }
