--- a/net/minecraft/world/level/block/FurnaceBlock.java
+++ b/net/minecraft/world/level/block/FurnaceBlock.java
@@ -45,8 +_,7 @@
     @Override
     protected void openContainer(Level p_53631_, BlockPos p_53632_, Player p_53633_) {
         BlockEntity blockentity = p_53631_.getBlockEntity(p_53632_);
-        if (blockentity instanceof FurnaceBlockEntity) {
-            p_53633_.openMenu((MenuProvider)blockentity);
+        if (blockentity instanceof FurnaceBlockEntity && p_53633_.openMenu((MenuProvider)blockentity).isPresent()) { // Paper - Fix InventoryOpenEvent cancellation
             p_53633_.awardStat(Stats.INTERACT_WITH_FURNACE);
         }
     }
