--- a/net/minecraft/world/level/block/HayBlock.java
+++ b/net/minecraft/world/level/block/HayBlock.java
@@ -23,6 +_,6 @@
 
     @Override
     public void fallOn(Level p_153362_, BlockState p_153363_, BlockPos p_153364_, Entity p_153365_, float p_153366_) {
-        p_153365_.causeFallDamage(p_153366_, 0.2F, p_153362_.damageSources().fall());
+        super.fallOn(p_153362_, p_153363_, p_153364_, p_153365_, p_153366_); // Purpur
     }
 }
