--- a/net/minecraft/world/level/block/HoneyBlock.java
+++ b/net/minecraft/world/level/block/HoneyBlock.java
@@ -60,6 +_,7 @@
 
     @Override
     protected void entityInside(BlockState p_54003_, Level p_54004_, BlockPos p_54005_, Entity p_54006_) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(p_54006_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_54004_, p_54005_)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (this.isSlidingDown(p_54005_, p_54006_)) {
             this.maybeDoSlideAchievement(p_54006_, p_54005_);
             this.doSlideMovement(p_54006_);
