--- a/net/minecraft/world/level/block/HopperBlock.java
+++ b/net/minecraft/world/level/block/HopperBlock.java
@@ -187,6 +_,7 @@
 
     @Override
     protected void entityInside(BlockState p_54066_, Level p_54067_, BlockPos p_54068_, Entity p_54069_) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(p_54069_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_54067_, p_54068_)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         BlockEntity blockentity = p_54067_.getBlockEntity(p_54068_);
         if (blockentity instanceof HopperBlockEntity) {
             HopperBlockEntity.entityInside(p_54067_, p_54068_, p_54066_, p_54069_, (HopperBlockEntity)blockentity);
