--- a/net/minecraft/world/level/block/IceBlock.java
+++ b/net/minecraft/world/level/block/IceBlock.java
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class IceBlock extends HalfTransparentBlock {
     public static final MapCodec<IceBlock> CODEC = simpleCodec(IceBlock::new);
@@ -47,6 +_,23 @@
         }
     }
 
+    public void afterDestroy(Level world, BlockPos pos, ItemStack tool) {
+        // Paper end - Improve Block#breakNaturally API
+        if (!EnchantmentHelper.hasTag(tool, EnchantmentTags.PREVENTS_ICE_MELTING)) {
+            if (world.dimensionType().ultraWarm()) {
+                world.removeBlock(pos, false);
+                return;
+            }
+
+            BlockState iblockdata1 = world.getBlockState(pos.below());
+
+            if (iblockdata1.blocksMotion() || iblockdata1.liquid()) {
+                world.setBlockAndUpdate(pos, IceBlock.meltsInto());
+            }
+        }
+
+    }
+
     @Override
     protected void randomTick(BlockState p_221355_, ServerLevel p_221356_, BlockPos p_221357_, RandomSource p_221358_) {
         if (p_221356_.getBrightness(LightLayer.BLOCK, p_221357_) > 11 - p_221355_.getLightBlock(p_221356_, p_221357_)) {
@@ -55,6 +_,11 @@
     }
 
     protected void melt(BlockState p_54169_, Level p_54170_, BlockPos p_54171_) {
+        // CraftBukkit start
+        if (CraftEventFactory.callBlockFadeEvent(p_54170_, p_54171_, p_54170_.dimensionType().ultraWarm() ? Blocks.AIR.defaultBlockState() : Blocks.WATER.defaultBlockState()).isCancelled()) {
+            return;
+        }
+        // CraftBukkit end
         if (p_54170_.dimensionType().ultraWarm()) {
             p_54170_.removeBlock(p_54171_, false);
         } else {
