--- a/net/minecraft/world/level/block/JigsawBlock.java
+++ b/net/minecraft/world/level/block/JigsawBlock.java
@@ -69,15 +_,13 @@
     }
 
     public static boolean canAttach(StructureTemplate.JigsawBlockInfo source, StructureTemplate.JigsawBlockInfo target) {
-        Direction sourceFront = getFrontFacing(source.state());
-        Direction targetFront = getFrontFacing(target.state());
-        Direction sourceTop = getTopFacing(source.state());
-        Direction targetTop = getTopFacing(target.state());
+        Direction sourceFront = getFrontFacing(source.info().state());
+        Direction targetFront = getFrontFacing(target.info().state());
+        Direction sourceTop = getTopFacing(source.info().state());
+        Direction targetTop = getTopFacing(target.info().state());
         JigsawBlockEntity.JointType jointType = source.jointType();
         boolean rollable = jointType == JigsawBlockEntity.JointType.ROLLABLE;
-        return sourceFront == targetFront.getOpposite()
-            && (rollable || sourceTop == targetTop)
-            && (target.name() == null || source.target().equals(target.name()));
+        return sourceFront == targetFront.getOpposite() && (rollable || sourceTop == targetTop) && source.target().equals(target.name());
     }
 
     public static Direction getFrontFacing(BlockState state) {
