--- a/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -2,7 +_,6 @@
 
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.cauldron.CauldronInteraction;
 import net.minecraft.world.entity.Entity;
@@ -16,6 +_,8 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.craftbukkit.block.CraftBlockStates;
 
 public class LayeredCauldronBlock extends AbstractCauldronBlock {
     public static final MapCodec<LayeredCauldronBlock> CODEC = RecordCodecBuilder.mapCodec(
@@ -61,11 +_,16 @@
 
     @Override
     protected void entityInside(BlockState p_153534_, Level p_153535_, BlockPos p_153536_, Entity p_153537_) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(p_153537_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_153535_, p_153536_)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (!p_153535_.isClientSide && p_153537_.isOnFire() && this.isEntityInsideContent(p_153534_, p_153536_, p_153537_)) {
+            // CraftBukkit start
+            if ((p_153537_ instanceof net.minecraft.world.entity.player.Player || p_153535_.getGameRules().getBoolean(net.minecraft.world.level.GameRules.RULE_MOBGRIEFING)) && p_153537_.mayInteract(p_153535_, p_153536_)) { // Paper - Fixes MC-248588
+                if (!lowerFillLevel(p_153534_, p_153535_, p_153536_, p_153537_, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.EXTINGUISH)) {
+                    return;
+                }
+            }
             p_153537_.clearFire();
-            if (p_153537_.mayInteract(p_153535_, p_153536_)) {
-                this.handleEntityOnFireInside(p_153534_, p_153535_, p_153536_);
-            }
+            // CraftBukkit end
         }
     }
 
@@ -78,18 +_,40 @@
     }
 
     public static void lowerFillLevel(BlockState p_153560_, Level p_153561_, BlockPos p_153562_) {
+        // CraftBukkit start
+        lowerFillLevel(p_153560_, p_153561_, p_153562_, null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.UNKNOWN);
+    }
+
+    public static boolean lowerFillLevel(BlockState p_153560_, Level p_153561_, BlockPos p_153562_, Entity entity, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason reason) {
         int i = p_153560_.getValue(LEVEL) - 1;
         BlockState blockstate = i == 0 ? Blocks.CAULDRON.defaultBlockState() : p_153560_.setValue(LEVEL, Integer.valueOf(i));
-        p_153561_.setBlockAndUpdate(p_153562_, blockstate);
-        p_153561_.gameEvent(GameEvent.BLOCK_CHANGE, p_153562_, GameEvent.Context.of(blockstate));
-    }
+        return changeLevel(p_153560_, p_153561_, p_153562_, blockstate, entity, reason);
+    }
+
+    // CraftBukkit start
+    public static boolean changeLevel(BlockState iblockdata, Level world, BlockPos blockposition, BlockState newBlock, Entity entity, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason reason) {
+        CraftBlockState newState = CraftBlockStates.getBlockState(world, blockposition);
+        newState.setData(newBlock);
+
+        org.bukkit.event.block.CauldronLevelChangeEvent event = new org.bukkit.event.block.CauldronLevelChangeEvent(
+                world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()),
+                (entity == null) ? null : entity.getBukkitEntity(), reason, newState
+        );
+        world.getCraftServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            return false;
+        }
+        newState.update(true);
+        world.gameEvent((net.minecraft.core.Holder<net.minecraft.world.level.gameevent.GameEvent>) GameEvent.BLOCK_CHANGE, blockposition, GameEvent.Context.of(newBlock));
+        return true;
+    }
+    // CraftBukkit end
 
     @Override
     public void handlePrecipitation(BlockState p_153539_, Level p_153540_, BlockPos p_153541_, Biome.Precipitation p_153542_) {
         if (CauldronBlock.shouldHandlePrecipitation(p_153540_, p_153542_) && p_153539_.getValue(LEVEL) != 3 && p_153542_ == this.precipitationType) {
             BlockState blockstate = p_153539_.cycle(LEVEL);
-            p_153540_.setBlockAndUpdate(p_153541_, blockstate);
-            p_153540_.gameEvent(GameEvent.BLOCK_CHANGE, p_153541_, GameEvent.Context.of(blockstate));
+            changeLevel(p_153539_, p_153540_, p_153541_, blockstate, null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
         }
     }
 
@@ -107,8 +_,11 @@
     protected void receiveStalactiteDrip(BlockState p_153544_, Level p_153545_, BlockPos p_153546_, Fluid p_153547_) {
         if (!this.isFull(p_153544_)) {
             BlockState blockstate = p_153544_.setValue(LEVEL, Integer.valueOf(p_153544_.getValue(LEVEL) + 1));
-            p_153545_.setBlockAndUpdate(p_153546_, blockstate);
-            p_153545_.gameEvent(GameEvent.BLOCK_CHANGE, p_153546_, GameEvent.Context.of(blockstate));
+            // CraftBukkit start
+            if (!changeLevel(p_153544_, p_153545_, p_153546_, blockstate, null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL)) {
+                return;
+            }
+            // CraftBukkit end
             p_153545_.levelEvent(1047, p_153546_, 0);
         }
     }
