--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -28,6 +_,8 @@
 
 import net.neoforged.neoforge.common.IShearable;
 
+import org.bukkit.event.block.LeavesDecayEvent; // CraftBukkit
+
 public class LeavesBlock extends Block implements SimpleWaterloggedBlock, IShearable {
     public static final int DECAY_DISTANCE = 7;
     public static final IntegerProperty DISTANCE = BlockStateProperties.DISTANCE;
@@ -65,6 +_,14 @@
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (this.decaying(state)) {
+            // CraftBukkit start
+            LeavesDecayEvent event = new LeavesDecayEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos));
+            level.getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled() || !level.getBlockState(pos).is(this)) {
+                return;
+            }
+            // CraftBukkit end
             dropResources(state, level, pos);
             level.removeBlock(pos, false);
         }
