--- a/net/minecraft/world/level/block/LevelEvent.java
+++ b/net/minecraft/world/level/block/LevelEvent.java
@@ -71,10 +_,7 @@
     public static final int PARTICLES_TURTLE_EGG_PLACEMENT = 2012;
     public static final int PARTICLES_SMASH_ATTACK = 2013;
     public static final int PARTICLES_DESTROY_BLOCK = 2014;
-    public static final int PARTICLES_DRAGON_EGG_TELEPORT = 2015;
-    public static final int PARTICLES_SHULKER_TELEPORT = 2016;
-    public static final int PARTICLES_CONSUME_EFFECT_TELEPORT = 2017;
-    public static final int PARTICLES_ENDERMAN_TELEPORT = 2018;
+    public static final int PARTICLES_DRAGON_EGG = 2015;
     public static final int ANIMATION_END_GATEWAY_SPAWN = 3000;
     public static final int ANIMATION_DRAGON_SUMMON_ROAR = 3001;
     public static final int PARTICLES_ELECTRIC_SPARK = 3002;
