--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -195,7 +_,7 @@
 
     @Override
     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block block, @Nullable Orientation orientation, boolean movedByPiston) {
-        if (this.shouldSpreadLiquid(level, pos, state)) {
+        if (!net.neoforged.neoforge.fluids.FluidInteractionRegistry.canInteract(level, pos)) {
             level.scheduleTick(pos, state.getFluidState().getType(), this.fluid.getTickDelay(level));
         }
 
@@ -211,6 +_,7 @@
         }
     }
 
+    @Deprecated // Neo: Use FluidInteractionRegistry#canInteract instead
     private boolean shouldSpreadLiquid(Level level, BlockPos pos, BlockState state) {
         if (this.fluid.is(FluidTags.LAVA)) {
             boolean isOverSoulSoil = level.getBlockState(pos.below()).is(Blocks.SOUL_SOIL);
@@ -219,14 +_,20 @@
                 BlockPos neighbourPos = pos.relative(direction.getOpposite());
                 if (level.getFluidState(neighbourPos).is(FluidTags.WATER)) {
                     Block convertToBlock = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
-                    level.setBlockAndUpdate(pos, convertToBlock.defaultBlockState());
-                    this.fizz(level, pos);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, convertToBlock.defaultBlockState(), Block.UPDATE_ALL)) {
+                        this.fizz(level, pos);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
 
                 if (isOverSoulSoil && level.getBlockState(neighbourPos).is(Blocks.BLUE_ICE)) {
-                    level.setBlockAndUpdate(pos, Blocks.BASALT.defaultBlockState());
-                    this.fizz(level, pos);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, Blocks.BASALT.defaultBlockState(), Block.UPDATE_ALL)) {
+                        this.fizz(level, pos);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
             }
