--- a/net/minecraft/world/level/block/MushroomBlock.java
+++ b/net/minecraft/world/level/block/MushroomBlock.java
@@ -20,6 +_,7 @@
 import net.minecraft.world.level.levelgen.feature.configurations.HugeMushroomFeatureConfiguration;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.TreeType;
 
 public class MushroomBlock extends VegetationBlock implements BonemealableBlock {
     public static final MapCodec<MushroomBlock> CODEC = RecordCodecBuilder.mapCodec(
@@ -69,7 +_,7 @@
             }
 
             if (level.isEmptyBlock(offset) && state.canSurvive(level, offset)) {
-                level.setBlock(offset, state, 2);
+                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, offset, state, 2); // CraftBukkit
             }
         }
     }
@@ -83,16 +_,26 @@
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
         BlockPos belowPos = pos.below();
         BlockState below = level.getBlockState(belowPos);
-        return below.is(BlockTags.OVERRIDES_MUSHROOM_LIGHT_REQUIREMENT) ? true : level.getRawBrightness(pos, 0) < 13 && this.mayPlaceOn(below, level, belowPos);
+        var soilDecision = below.canSustainPlant(level, belowPos, net.minecraft.core.Direction.UP, state);
+        return below.is(BlockTags.OVERRIDES_MUSHROOM_LIGHT_REQUIREMENT) ? true : soilDecision.isDefault() ? (level.getRawBrightness(pos, 0) < 13 && this.mayPlaceOn(below, level, belowPos)) : soilDecision.isTrue();
     }
 
     public boolean growMushroom(ServerLevel level, BlockPos pos, BlockState state, RandomSource random) {
         Optional<? extends Holder<ConfiguredFeature<?, ?>>> feature = level.registryAccess().lookupOrThrow(Registries.CONFIGURED_FEATURE).get(this.feature);
+
+        // Neo: Fire the BlockGrowFeatureEvent and update the result of the Optional local with the new feature.
+        var event = net.neoforged.neoforge.event.EventHooks.fireBlockGrowFeature(level, random, pos, feature.orElse(null));
+        if (event.isCanceled()) {
+            return false;
+        }
+        feature = Optional.ofNullable(event.getFeature());
+
         if (feature.isEmpty()) {
             return false;
         }
 
         level.removeBlock(pos, false);
+        SaplingBlock.treeType = (this == Blocks.BROWN_MUSHROOM) ? TreeType.BROWN_MUSHROOM : TreeType.RED_MUSHROOM; // CraftBukkit
         if (feature.get().value().place(level, level.getChunkSource().getGenerator(), random, pos)) {
             return true;
         }
