--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -69,7 +_,7 @@
 
     @Override
     protected void randomTick(BlockState p_221799_, ServerLevel p_221800_, BlockPos p_221801_, RandomSource p_221802_) {
-        if (p_221800_.dimensionType().natural()
+        if (p_221800_.spigotConfig.enableZombiePigmenPortalSpawns && p_221800_.dimensionType().natural() // Spigot
             && p_221800_.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)
             && p_221802_.nextInt(2000) < p_221800_.getDifficulty().getId()) {
             while (p_221800_.getBlockState(p_221801_).is(this)) {
@@ -97,6 +_,7 @@
 
     @Override
     protected void entityInside(BlockState p_54915_, Level p_54916_, BlockPos p_54917_, Entity p_54918_) {
+        if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(p_54918_.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(p_54916_, p_54917_)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (p_54918_.canUsePortal(false)) {
             p_54918_.setAsInsidePortal(this, p_54917_);
         }
