--- a/net/minecraft/world/level/block/NoteBlock.java
+++ b/net/minecraft/world/level/block/NoteBlock.java
@@ -57,6 +_,7 @@
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return this.defaultBlockState(); // Paper - place without considering instrument
         return this.setInstrument(context.getLevel(), context.getClickedPos(), this.defaultBlockState());
     }
 
@@ -71,6 +_,7 @@
         BlockState neighbourState,
         RandomSource random
     ) {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return state; // Paper - prevent noteblock instrument from updating
         boolean neighborDirectionSetsInstrument = directionToNeighbour.getAxis() == Direction.Axis.Y;
         return neighborDirectionSetsInstrument
             ? this.setInstrument(level, pos, state)
@@ -79,18 +_,25 @@
 
     @Override
     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block block, @Nullable Orientation orientation, boolean movedByPiston) {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) return; // Paper - prevent noteblock powered-state from updating
         boolean signal = level.hasNeighborSignal(pos);
         if (signal != state.getValue(POWERED)) {
+            // Paper start - Call BlockRedstoneEvent
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBinaryRedstoneChange(level, pos, signal)) {
+                return;
+            }
+            // Paper end - Call BlockRedstoneEvent
             if (signal) {
                 this.playNote(null, state, level, pos);
+                state = level.getBlockState(pos); // CraftBukkit - SPIGOT-5617: update in case changed in event
             }
 
-            level.setBlockAndUpdate(pos, state.setValue(POWERED, signal));
+            level.setBlock(pos, state.setValue(POWERED, signal), 3);
         }
     }
 
     private void playNote(@Nullable Entity source, BlockState state, Level level, BlockPos pos) {
-        if (state.getValue(INSTRUMENT).worksAboveNoteBlock() || level.getBlockState(pos.above()).isAir()) {
+        if (level.purpurConfig.noteBlockIgnoreAbove || state.getValue(INSTRUMENT).worksAboveNoteBlock() || level.getBlockState(pos.above()).isAir()) { // Purpur - Config to allow Note Block sounds when blocked
             level.blockEvent(pos, this, 0, 0);
             level.gameEvent(source, GameEvent.NOTE_BLOCK_PLAY, pos);
         }
@@ -108,7 +_,9 @@
     @Override
     protected InteractionResult useWithoutItem(BlockState state, Level level, BlockPos pos, Player player, BlockHitResult hitResult) {
         if (!level.isClientSide()) {
-            state = state.cycle(NOTE);
+            int _new = net.neoforged.neoforge.common.CommonHooks.onNoteChange(level, pos, state, state.getValue(NOTE), state.cycle(NOTE).getValue(NOTE));
+            if (_new == -1) return InteractionResult.FAIL;
+            if (!io.papermc.paper.configuration.GlobalConfiguration.get().blockUpdates.disableNoteblockUpdates) state = state.setValue(NOTE, _new); // Paper - prevent noteblock note from updating
             level.setBlockAndUpdate(pos, state);
             this.playNote(player, state, level, pos);
             player.awardStat(Stats.TUNE_NOTEBLOCK);
@@ -131,10 +_,17 @@
 
     @Override
     protected boolean triggerEvent(BlockState state, Level level, BlockPos pos, int b0, int b1) {
+        net.neoforged.neoforge.event.level.NoteBlockEvent.Play e = new net.neoforged.neoforge.event.level.NoteBlockEvent.Play(level, pos, state, state.getValue(NOTE), state.getValue(INSTRUMENT));
+        if (net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(e).isCanceled()) return false;
+        state = state.setValue(NOTE, e.getVanillaNoteId()).setValue(INSTRUMENT, e.getInstrument());
         NoteBlockInstrument instrument = state.getValue(INSTRUMENT);
+        // Paper start - move NotePlayEvent call to fix instrument/note changes
+        org.bukkit.event.block.NotePlayEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callNotePlayEvent(level, pos, instrument, state.getValue(NOTE));
+        if (event.isCancelled()) return false;
+        // Paper end - move NotePlayEvent call to fix instrument/note changes
         float pitch;
         if (instrument.isTunable()) {
-            int note = state.getValue(NOTE);
+            int note = event.getNote().getId(); // Paper - move NotePlayEvent call to fix instrument/note changes
             pitch = getPitchFromNote(note);
             level.addParticle(ParticleTypes.NOTE, pos.getX() + 0.5, pos.getY() + 1.2, pos.getZ() + 0.5, note / 24.0, 0.0, 0.0);
         } else {
@@ -150,7 +_,7 @@
 
             soundEvent = Holder.direct(SoundEvent.createVariableRangeEvent(soundId));
         } else {
-            soundEvent = instrument.getSoundEvent();
+            soundEvent = org.bukkit.craftbukkit.block.data.CraftBlockData.toVanilla(event.getInstrument(), NoteBlockInstrument.class).getSoundEvent(); // Paper - move NotePlayEvent call to fix instrument/note changes
         }
 
         level.playSeededSound(
