--- a/net/minecraft/world/level/block/PathBlock.java
+++ b/net/minecraft/world/level/block/PathBlock.java
@@ -22,7 +_,7 @@
     private static final VoxelShape SHAPE = Block.column(16.0, 0.0, 15.0);
     private final Block baseBlock;
 
-    public PathBlock(Block baseBlock, BlockBehaviour.Properties properties) {
+    protected PathBlock(Block baseBlock, BlockBehaviour.Properties properties) {
         super(properties);
         this.baseBlock = baseBlock;
     }
