--- a/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -9,7 +_,6 @@
 import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.tags.FluidTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.attribute.EnvironmentAttributes;
 import net.minecraft.world.entity.Entity;
@@ -53,7 +_,7 @@
     @Override
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, double fallDistance) {
         if (state.getValue(TIP_DIRECTION) == Direction.UP && state.getValue(THICKNESS) == SpeleothemThickness.TIP) {
-            entity.causeFallDamage(fallDistance + 2.5, 2.0F, level.damageSources().stalagmite());
+            entity.causeFallDamage(fallDistance + 2.5, 2.0F, level.damageSources().stalagmite().eventBlockDamager(level, pos)); // CraftBukkit
         } else {
             super.fallOn(level, state, pos, entity, fallDistance);
         }
@@ -79,7 +_,7 @@
 
     @VisibleForTesting
     public static void maybeTransferFluid(BlockState state, ServerLevel level, BlockPos pos, float randomValue) {
-        if (!(randomValue > 0.17578125F) || !(randomValue > 0.05859375F)) {
+        if (true) { //Neo: remove the water and lava drip chance checks to allow modded fluids to drip into cauldrons
             if (isStalactiteStartPos(state, level, pos)) {
                 Optional<PointedDripstoneBlock.FluidInfo> fluidInfo = getFluidAboveStalactite(level, pos, state);
                 if (!fluidInfo.isEmpty()) {
@@ -88,22 +_,21 @@
                     if (fluid == Fluids.WATER) {
                         transferProbability = 0.17578125F;
                     } else {
-                        if (fluid != Fluids.LAVA) {
-                            return;
-                        }
 
                         transferProbability = 0.05859375F;
                     }
 
-                    if (!(randomValue >= transferProbability)) {
+                    net.neoforged.neoforge.fluids.FluidType.DripstoneDripInfo dripInfo = fluid.getFluidType().getDripInfo();
+                    if (dripInfo != null && !(randomValue >= dripInfo.chance())) {
                         BlockPos stalactiteTipPos = findTip(state, level, pos, 11, false);
                         if (stalactiteTipPos != null) {
                             if (fluidInfo.get().sourceState.is(Blocks.MUD) && fluid == Fluids.WATER) {
                                 BlockState newState = Blocks.CLAY.defaultBlockState();
-                                level.setBlockAndUpdate(fluidInfo.get().pos, newState);
+                                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, fluidInfo.get().pos, newState, Block.UPDATE_ALL)) { // Paper - Call BlockFormEvent
                                 Block.pushEntitiesUp(fluidInfo.get().sourceState, newState, level, fluidInfo.get().pos);
                                 level.gameEvent(GameEvent.BLOCK_CHANGE, fluidInfo.get().pos, GameEvent.Context.of(newState));
                                 level.levelEvent(1504, stalactiteTipPos, 0);
+                                } // Paper - Call BlockFormEvent
                             } else {
                                 BlockPos cauldronPos = findFillableCauldronBelowStalactiteTip(level, stalactiteTipPos, fluid);
                                 if (cauldronPos != null) {
@@ -180,7 +_,7 @@
     }
 
     private static boolean canFillCauldron(Fluid fluidAbove) {
-        return fluidAbove == Fluids.LAVA || fluidAbove == Fluids.WATER;
+        return fluidAbove.getFluidType().getDripInfo() != null;
     }
 
     @Override
@@ -193,7 +_,9 @@
         if (fluidAbove.isSame(Fluids.EMPTY)) {
             return level.environmentAttributes().getValue(EnvironmentAttributes.DEFAULT_DRIPSTONE_PARTICLE, posAbove);
         } else {
-            return fluidAbove.is(FluidTags.LAVA) ? ParticleTypes.DRIPPING_DRIPSTONE_LAVA : ParticleTypes.DRIPPING_DRIPSTONE_WATER;
+            ParticleOptions options = fluidAbove.getFluidType().getDripInfo() != null ? fluidAbove.getFluidType().getDripInfo().dripParticle() : ParticleTypes.DRIPPING_DRIPSTONE_WATER;
+            if (options == null) options = level.environmentAttributes().getValue(EnvironmentAttributes.DEFAULT_DRIPSTONE_PARTICLE, posAbove);
+            return options;
         }
     }
 
