--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -240,7 +_,7 @@
         BlockState relativeState = level.getBlockState(relativePos);
         if (canConnectUp) {
             boolean isPlaceableAbove = relativeState.getBlock() instanceof TrapDoorBlock || this.canSurviveOn(level, relativePos, relativeState);
-            if (isPlaceableAbove && shouldConnectTo(level.getBlockState(relativePos.above()))) {
+            if (isPlaceableAbove && level.getBlockState(relativePos.above()).canRedstoneConnectTo(level, relativePos.above(), null)) {
                 if (relativeState.isFaceSturdy(level, relativePos, direction.getOpposite())) {
                     return RedstoneSide.UP;
                 }
@@ -249,10 +_,14 @@
             }
         }
 
-        return !shouldConnectTo(relativeState, direction)
-                && (relativeState.isRedstoneConductor(level, relativePos) || !shouldConnectTo(level.getBlockState(relativePos.below())))
-            ? RedstoneSide.NONE
-            : RedstoneSide.SIDE;
+        if (relativeState.canRedstoneConnectTo(level, relativePos, direction)) {
+             return RedstoneSide.SIDE;
+        } else if (relativeState.isRedstoneConductor(level, relativePos)) {
+             return RedstoneSide.NONE;
+        } else {
+             BlockPos blockPosBelow = relativePos.below();
+             return level.getBlockState(blockPosBelow).canRedstoneConnectTo(level, blockPosBelow, null) ? RedstoneSide.SIDE : RedstoneSide.NONE;
+        }
     }
 
     @Override
@@ -266,6 +_,56 @@
         return relativeState.isFaceSturdy(level, relativePos, Direction.UP) || relativeState.is(Blocks.HOPPER);
     }
 
+    // Paper start - Optimize redstone (Eigencraft)
+    // The bulk of the new functionality is found in RedstoneWireTurbo.java
+    // io.papermc.paper.redstone.RedstoneWireTurbo turbo = new io.papermc.paper.redstone.RedstoneWireTurbo(this);
+
+    /*
+     * Modified version of pre-existing updateSurroundingRedstone, which is called from
+     * this.neighborChanged and a few other methods in this class.
+     * Note: Added 'source' argument so as to help determine direction of information flow
+     */
+    private void updateSurroundingRedstone(Level worldIn, BlockPos pos, BlockState state, @Nullable Orientation orientation, boolean blockAdded) {
+        if (worldIn.paperConfig().misc.redstoneImplementation == io.papermc.paper.configuration.WorldConfiguration.Misc.RedstoneImplementation.EIGENCRAFT) {
+            // since 24w33a the source pos is no longer given, but instead an Orientation parameter
+            // when this is not null, it can be used to find the source pos, which the turbo uses
+            // to find the direction of information flow
+            BlockPos source = null;
+            if (orientation != null) {
+                source = pos.relative(orientation.getFront().getOpposite());
+            }
+            worldIn.turbo.updateSurroundingRedstone(worldIn, pos, state, source); // SparklyPaper - parallel world ticking
+            return;
+        }
+        updatePowerStrength(worldIn, pos, state, orientation, blockAdded);
+    }
+
+    /*
+     * This method computes a wire's target strength and updates the given block state.
+     * It uses the DefaultRedstoneWireEvaluator for this, which is identical to code
+     * that was present in this class prior to the introduction of the experimental redstone
+     * changes in 24w33a.
+     * The previous implementation of this method in this patch had optimizations that have
+     * not been relevant since 1.13, thus it has been greatly simplified.
+     */
+    public BlockState calculateCurrentChanges(Level level, BlockPos pos, BlockState state) {
+        int oldPower = state.getValue(POWER);
+        int newPower = ((DefaultRedstoneWireEvaluator) evaluator).calculateTargetStrength(level, pos);
+        if (oldPower != newPower) {
+            if (level.getBlockState(pos) == state) {
+                newPower = org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, pos, oldPower, newPower).getNewCurrent();
+                state = state.setValue(POWER, newPower);
+                // [Space Walker] suppress shape updates and emit those manually to
+                // bypass the new neighbor update stack.
+                if (level.setBlock(pos, state, Block.UPDATE_KNOWN_SHAPE | Block.UPDATE_CLIENTS)) {
+                    level.turbo.updateNeighborShapes(level, pos, state); // SparklyPaper - parallel world ticking
+                }
+            }
+        }
+        return state;
+    }
+    // Paper end
+
     private void updatePowerStrength(
         Level level, BlockPos pos, BlockState state, @Nullable Orientation orientation, boolean shapeUpdateWiresAroundInitialPosition
     ) {
@@ -296,7 +_,13 @@
     @Override
     protected void onPlace(BlockState state, Level level, BlockPos pos, BlockState oldState, boolean movedByPiston) {
         if (!oldState.is(state.getBlock()) && !level.isClientSide()) {
-            this.updatePowerStrength(level, pos, state, null, true);
+            // Paper start - optimize redstone - replace call to updatePowerStrength
+            if (level.paperConfig().misc.redstoneImplementation == io.papermc.paper.configuration.WorldConfiguration.Misc.RedstoneImplementation.ALTERNATE_CURRENT) {
+                level.getWireHandler().onWireAdded(pos, state); // Alternate Current
+            } else {
+                this.updateSurroundingRedstone(level, pos, state, null, true); // Vanilla/Eigencraft
+            }
+            // Paper end - optimize redstone
 
             for (Direction direction : Direction.Plane.VERTICAL) {
                 level.updateNeighborsAt(pos.relative(direction), this);
@@ -313,7 +_,13 @@
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
 
-            this.updatePowerStrength(level, pos, state, null, false);
+            // Paper start - optimize redstone - replace call to updatePowerStrength
+            if (level.paperConfig().misc.redstoneImplementation == io.papermc.paper.configuration.WorldConfiguration.Misc.RedstoneImplementation.ALTERNATE_CURRENT) {
+                level.getWireHandler().onWireRemoved(pos, state); // Alternate Current
+            } else {
+                this.updateSurroundingRedstone(level, pos, state, null, false); // Vanilla/Eigencraft
+            }
+            // Paper end - optimize redstone
             this.updateNeighborsOfNeighboringWires(level, pos);
         }
     }
@@ -336,14 +_,20 @@
     @Override
     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block block, @Nullable Orientation orientation, boolean movedByPiston) {
         if (!level.isClientSide()) {
-            if (block != this || !useExperimentalEvaluator(level)) {
-                if (state.canSurvive(level, pos)) {
-                    this.updatePowerStrength(level, pos, state, orientation, false);
-                } else {
-                    dropResources(state, level, pos);
-                    level.removeBlock(pos, false);
+            // Paper start - optimize redstone (Alternate Current)
+            // Alternate Current handles breaking of redstone wires in the WireHandler.
+            if (level.paperConfig().misc.redstoneImplementation == io.papermc.paper.configuration.WorldConfiguration.Misc.RedstoneImplementation.ALTERNATE_CURRENT) {
+                level.getWireHandler().onWireUpdated(pos, state, orientation);
+            } else
+                // Paper end - optimize redstone (Alternate Current)
+                if (block != this || !useExperimentalEvaluator(level)) {
+                    if (state.canSurvive(level, pos)) {
+                        this.updateSurroundingRedstone(level, pos, state, orientation, false); // Paper - Optimize redstone (Eigencraft)
+                    } else {
+                        dropResources(state, level, pos);
+                        level.removeBlock(pos, false);
+                    }
                 }
-            }
         }
     }
 
