--- a/net/minecraft/world/level/block/RepeaterBlock.java
+++ b/net/minecraft/world/level/block/RepeaterBlock.java
@@ -7,7 +_,6 @@
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.context.BlockPlaceContext;
-import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.ScheduledTickAccess;
@@ -18,7 +_,6 @@
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
 import net.minecraft.world.phys.BlockHitResult;
-import org.jspecify.annotations.Nullable;
 
 public class RepeaterBlock extends DiodeBlock {
     public static final BooleanProperty LOCKED = BlockStateProperties.LOCKED;
@@ -70,12 +_,6 @@
                 ? state.setValue(LOCKED, this.isLocked(level, pos, state))
                 : super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
         }
-    }
-
-    @Override
-    protected boolean shouldRedstoneWireConnectTo(BlockState state, BlockGetter level, BlockPos pos, @Nullable Direction direction) {
-        Direction repeaterDirection = state.getValue(FACING);
-        return repeaterDirection == direction || repeaterDirection.getOpposite() == direction;
     }
 
     @Override
