--- a/net/minecraft/world/level/block/RootedDirtBlock.java
+++ b/net/minecraft/world/level/block/RootedDirtBlock.java
@@ -8,6 +_,7 @@
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class RootedDirtBlock extends Block implements BonemealableBlock {
     public static final MapCodec<RootedDirtBlock> CODEC = simpleCodec(RootedDirtBlock::new);
@@ -33,7 +_,7 @@
 
     @Override
     public void performBonemeal(ServerLevel p_221974_, RandomSource p_221975_, BlockPos p_221976_, BlockState p_221977_) {
-        p_221974_.setBlockAndUpdate(p_221976_.below(), Blocks.HANGING_ROOTS.defaultBlockState());
+        CraftEventFactory.handleBlockSpreadEvent(p_221974_, p_221976_, p_221976_.below(), Blocks.HANGING_ROOTS.defaultBlockState()); // CraftBukkit
     }
 
     @Override
