--- a/net/minecraft/world/level/block/SculkSpreader.java
+++ b/net/minecraft/world/level/block/SculkSpreader.java
@@ -193,8 +_,8 @@
                     .forEach(p_222338_::add)
         );
         public static final int MAX_CURSOR_DECAY_DELAY = 1;
-        public BlockPos pos;
-        public int charge;
+        private BlockPos pos;
+        int charge;
         private int updateDelay;
         private int decayDelay;
         @Nullable
