--- a/net/minecraft/world/level/block/SelectableSlotContainer.java
+++ b/net/minecraft/world/level/block/SelectableSlotContainer.java
@@ -43,7 +_,7 @@
         };
     }
 
-    public static int getSection(float relativeCoordinate, int maxSections) {
+    private static int getSection(float relativeCoordinate, int maxSections) {
         float targetedPixel = relativeCoordinate * 16.0F;
         float sectionSize = 16.0F / maxSections;
         return Mth.clamp(Mth.floor(targetedPixel / sectionSize), 0, maxSections - 1);
