--- a/net/minecraft/world/level/block/ShelfBlock.java
+++ b/net/minecraft/world/level/block/ShelfBlock.java
@@ -102,6 +_,11 @@
         if (!level.isClientSide()) {
             boolean signal = level.hasNeighborSignal(pos);
             if (state.getValue(POWERED) != signal) {
+                // Paper start - Call BlockRedstoneEvent
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBinaryRedstoneChange(level, pos, signal)) {
+                    return;
+                }
+                // Paper end - Call BlockRedstoneEvent
                 BlockState newState = state.setValue(POWERED, signal);
                 if (!signal) {
                     newState = newState.setValue(SIDE_CHAIN_PART, SideChainPart.UNCONNECTED);
@@ -155,7 +_,7 @@
 
             Inventory inventory = player.getInventory();
             if (level.isClientSide()) {
-                return inventory.getSelectedItem().isEmpty() ? InteractionResult.PASS : InteractionResult.CONSUME;
+                return inventory.getSelectedItem().isEmpty() ? InteractionResult.PASS : InteractionResult.SUCCESS;
             }
 
             if (!state.getValue(POWERED)) {
@@ -170,7 +_,7 @@
                     this.playSound(level, pos, SoundEvents.SHELF_PLACE_ITEM);
                 }
 
-                return InteractionResult.SUCCESS_SERVER.heldItemTransformedTo(itemStack);
+                return InteractionResult.SUCCESS.heldItemTransformedTo(itemStack);
             } else {
                 ItemStack previousItem = inventory.getSelectedItem();
                 boolean anySwapped = this.swapHotbar(level, pos, inventory);
@@ -180,8 +_,8 @@
 
                 this.playSound(level, pos, SoundEvents.SHELF_MULTI_SWAP);
                 return previousItem == inventory.getSelectedItem()
-                    ? InteractionResult.SUCCESS_SERVER
-                    : InteractionResult.SUCCESS_SERVER.heldItemTransformedTo(inventory.getSelectedItem());
+                    ? InteractionResult.SUCCESS
+                    : InteractionResult.SUCCESS.heldItemTransformedTo(inventory.getSelectedItem());
             }
         } else {
             return InteractionResult.PASS;
