--- a/net/minecraft/world/level/block/ShulkerBoxBlock.java
+++ b/net/minecraft/world/level/block/ShulkerBoxBlock.java
@@ -73,7 +_,7 @@
     public static final EnumProperty<Direction> FACING = DirectionalBlock.FACING;
     public static final ResourceLocation CONTENTS = ResourceLocation.withDefaultNamespace("contents");
     @Nullable
-    public final DyeColor color;
+    private final DyeColor color;
 
     @Override
     public MapCodec<ShulkerBoxBlock> codec() {
