--- a/net/minecraft/world/level/block/SideChainPartBlock.java
+++ b/net/minecraft/world/level/block/SideChainPartBlock.java
@@ -97,11 +_,11 @@
     private void setPart(LevelAccessor level, BlockPos pos, SideChainPart newPart) {
         BlockState state = level.getBlockState(pos);
         if (this.getSideChainPart(state) != newPart) {
-            level.setBlockAndUpdate(pos, this.setSideChainPart(state, newPart));
+            level.setBlock(pos, this.setSideChainPart(state, newPart), 3);
         }
     }
 
-    record EmptyNeighbor(BlockPos pos) implements SideChainPartBlock.Neighbor {
+    public record EmptyNeighbor(BlockPos pos) implements SideChainPartBlock.Neighbor {
         @Override
         public boolean isConnectable() {
             return false;
@@ -118,7 +_,7 @@
         }
     }
 
-    sealed interface Neighbor permits SideChainPartBlock.EmptyNeighbor, SideChainPartBlock.SideChainNeighbor {
+    public sealed interface Neighbor permits SideChainPartBlock.EmptyNeighbor, SideChainPartBlock.SideChainNeighbor {
         BlockPos pos();
 
         boolean isConnectable();
@@ -140,7 +_,7 @@
         }
     }
 
-    record Neighbors(SideChainPartBlock block, LevelAccessor level, Direction facing, BlockPos center, Map<BlockPos, SideChainPartBlock.Neighbor> cache) {
+    public record Neighbors(SideChainPartBlock block, LevelAccessor level, Direction facing, BlockPos center, Map<BlockPos, SideChainPartBlock.Neighbor> cache) {
         private boolean isConnectableToThisBlock(BlockState neighbor) {
             return this.block.isConnectable(neighbor) && this.block.getFacing(neighbor) == this.facing;
         }
@@ -172,7 +_,7 @@
         }
     }
 
-    record SideChainNeighbor(LevelAccessor level, SideChainPartBlock block, BlockPos pos, SideChainPart part) implements SideChainPartBlock.Neighbor {
+    public record SideChainNeighbor(LevelAccessor level, SideChainPartBlock block, BlockPos pos, SideChainPart part) implements SideChainPartBlock.Neighbor {
         @Override
         public boolean isConnectable() {
             return true;
