--- a/net/minecraft/world/level/block/SignBlock.java
+++ b/net/minecraft/world/level/block/SignBlock.java
@@ -3,12 +_,12 @@
 import com.mojang.serialization.MapCodec;
 import java.util.Arrays;
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.network.chat.CommonComponents;
-import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.contents.PlainTextContents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
@@ -38,6 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public abstract class SignBlock extends BaseEntityBlock implements SimpleWaterloggedBlock {
     public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;
@@ -127,6 +_,7 @@
             } else if (!this.otherPlayerIsEditingSign(p_316173_, signblockentity)
                 && p_316173_.mayBuild()
                 && this.hasEditableText(p_316173_, signblockentity, flag1)) {
+                openTextEdit$cause.set(org.bukkit.event.player.PlayerSignOpenEvent.Cause.INTERACT);  // CraftBukkit // Youer
                 this.openTextEdit(p_316173_, signblockentity, flag1);
                 return InteractionResult.SUCCESS;
             } else {
@@ -169,11 +_,21 @@
         return woodtype;
     }
 
+    public AtomicReference<org.bukkit.event.player.PlayerSignOpenEvent.Cause> openTextEdit$cause = new AtomicReference<>(org.bukkit.event.player.PlayerSignOpenEvent.Cause.UNKNOWN);
     public void openTextEdit(Player p_277738_, SignBlockEntity p_277467_, boolean p_277771_) {
+        if (!CraftEventFactory.callPlayerSignOpenEvent(p_277738_, p_277467_, p_277771_, openTextEdit$cause.getAndSet(org.bukkit.event.player.PlayerSignOpenEvent.Cause.UNKNOWN))) {
+            return;
+        }
+        // CraftBukkit end
         p_277467_.setAllowedPlayerEditor(p_277738_.getUUID());
         p_277738_.openTextEdit(p_277467_, p_277771_);
     }
 
+    public void openTextEdit(Player p_277738_, SignBlockEntity p_277467_, boolean p_277771_, org.bukkit.event.player.PlayerSignOpenEvent.Cause cause) {
+        openTextEdit$cause.set(cause);
+        openTextEdit(p_277738_, p_277467_, p_277771_);
+    }
+
     private boolean otherPlayerIsEditingSign(Player p_277952_, SignBlockEntity p_277599_) {
         UUID uuid = p_277599_.getPlayerWhoMayEdit();
         return uuid != null && !uuid.equals(p_277952_.getUUID());
@@ -182,6 +_,6 @@
     @Nullable
     @Override
     public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level p_277367_, BlockState p_277896_, BlockEntityType<T> p_277724_) {
-        return createTickerHelper(p_277724_, BlockEntityType.SIGN, SignBlockEntity::tick);
+        return null; // CraftBukkit - remove unnecessary sign ticking
     }
 }
