--- a/net/minecraft/world/level/block/SignBlock.java
+++ b/net/minecraft/world/level/block/SignBlock.java
@@ -8,7 +_,6 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.network.chat.CommonComponents;
-import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.contents.PlainTextContents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
@@ -38,6 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public abstract class SignBlock extends BaseEntityBlock implements SimpleWaterloggedBlock {
     public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;
@@ -127,7 +_,7 @@
             } else if (!this.otherPlayerIsEditingSign(p_316173_, signblockentity)
                 && p_316173_.mayBuild()
                 && this.hasEditableText(p_316173_, signblockentity, flag1)) {
-                this.openTextEdit(p_316173_, signblockentity, flag1);
+                this.openTextEdit(p_316173_, signblockentity, flag1, org.bukkit.event.player.PlayerSignOpenEvent.Cause.INTERACT); // CraftBukkit
                 return InteractionResult.SUCCESS;
             } else {
                 return InteractionResult.PASS;
@@ -170,6 +_,15 @@
     }
 
     public void openTextEdit(Player p_277738_, SignBlockEntity p_277467_, boolean p_277771_) {
+        // CraftBukkit start
+        openTextEdit(p_277738_, p_277467_, p_277771_, org.bukkit.event.player.PlayerSignOpenEvent.Cause.UNKNOWN);
+    }
+
+    public void openTextEdit(Player p_277738_, SignBlockEntity p_277467_, boolean p_277771_, org.bukkit.event.player.PlayerSignOpenEvent.Cause cause) {
+        if (!CraftEventFactory.callPlayerSignOpenEvent(p_277738_, p_277467_, p_277771_, cause)) {
+            return;
+        }
+        // CraftBukkit end
         p_277467_.setAllowedPlayerEditor(p_277738_.getUUID());
         p_277738_.openTextEdit(p_277467_, p_277771_);
     }
@@ -182,6 +_,6 @@
     @Nullable
     @Override
     public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level p_277367_, BlockState p_277896_, BlockEntityType<T> p_277724_) {
-        return createTickerHelper(p_277724_, BlockEntityType.SIGN, SignBlockEntity::tick);
+        return null; // CraftBukkit - remove unnecessary sign ticking
     }
 }
