--- a/net/minecraft/world/level/block/SignBlock.java
+++ b/net/minecraft/world/level/block/SignBlock.java
@@ -3,12 +_,12 @@
 import com.mojang.serialization.MapCodec;
 import java.util.Arrays;
 import java.util.UUID;
+import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.network.chat.CommonComponents;
-import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.contents.PlainTextContents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.stats.Stats;
@@ -127,6 +_,7 @@
             } else if (!this.otherPlayerIsEditingSign(p_316173_, signblockentity)
                 && p_316173_.mayBuild()
                 && this.hasEditableText(p_316173_, signblockentity, flag1)) {
+                openTextEdit$cause.set(io.papermc.paper.event.player.PlayerOpenSignEvent.Cause.INTERACT);  // CraftBukkit // Youer
                 this.openTextEdit(p_316173_, signblockentity, flag1);
                 return InteractionResult.SUCCESS;
             } else {
@@ -169,11 +_,39 @@
         return woodtype;
     }
 
+    public AtomicReference<io.papermc.paper.event.player.PlayerOpenSignEvent.Cause> openTextEdit$cause = new AtomicReference<>(io.papermc.paper.event.player.PlayerOpenSignEvent.Cause.UNKNOWN);
     public void openTextEdit(Player p_277738_, SignBlockEntity p_277467_, boolean p_277771_) {
+        org.bukkit.entity.Player bukkitPlayer = (org.bukkit.entity.Player) p_277738_.getBukkitEntity();
+        org.bukkit.block.Block bukkitBlock = org.bukkit.craftbukkit.block.CraftBlock.at(p_277467_.getLevel(), p_277467_.getBlockPos());
+        org.bukkit.craftbukkit.block.CraftSign<?> bukkitSign = (org.bukkit.craftbukkit.block.CraftSign<?>) org.bukkit.craftbukkit.block.CraftBlockStates.getBlockState(bukkitBlock);
+        io.papermc.paper.event.player.PlayerOpenSignEvent.Cause cause = openTextEdit$cause.getAndSet(io.papermc.paper.event.player.PlayerOpenSignEvent.Cause.UNKNOWN);
+        io.papermc.paper.event.player.PlayerOpenSignEvent event = new io.papermc.paper.event.player.PlayerOpenSignEvent(
+                bukkitPlayer,
+                bukkitSign,
+                p_277771_ ? org.bukkit.block.sign.Side.FRONT : org.bukkit.block.sign.Side.BACK,
+                cause);
+        if (!event.callEvent()) return;
+        if (org.bukkit.event.player.PlayerSignOpenEvent.getHandlerList().getRegisteredListeners().length > 0) {
+            final org.bukkit.event.player.PlayerSignOpenEvent.Cause legacyCause = switch (cause) {
+                case PLACE -> org.bukkit.event.player.PlayerSignOpenEvent.Cause.PLACE;
+                case PLUGIN -> org.bukkit.event.player.PlayerSignOpenEvent.Cause.PLUGIN;
+                case INTERACT -> org.bukkit.event.player.PlayerSignOpenEvent.Cause.INTERACT;
+                case UNKNOWN -> org.bukkit.event.player.PlayerSignOpenEvent.Cause.UNKNOWN;
+            };
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerSignOpenEvent(p_277738_, p_277467_, p_277771_, legacyCause)) {
+                // Paper end - Add PlayerOpenSignEvent
+                return;
+            }
+        } // Paper - Add PlayerOpenSignEvent
         p_277467_.setAllowedPlayerEditor(p_277738_.getUUID());
         p_277738_.openTextEdit(p_277467_, p_277771_);
     }
 
+    public void openTextEdit(Player p_277738_, SignBlockEntity p_277467_, boolean p_277771_, io.papermc.paper.event.player.PlayerOpenSignEvent.Cause cause) {
+        openTextEdit$cause.set(cause);
+        openTextEdit(p_277738_, p_277467_, p_277771_);
+    }
+
     private boolean otherPlayerIsEditingSign(Player p_277952_, SignBlockEntity p_277599_) {
         UUID uuid = p_277599_.getPlayerWhoMayEdit();
         return uuid != null && !uuid.equals(p_277952_.getUUID());
@@ -182,6 +_,6 @@
     @Nullable
     @Override
     public <T extends BlockEntity> BlockEntityTicker<T> getTicker(Level p_277367_, BlockState p_277896_, BlockEntityType<T> p_277724_) {
-        return createTickerHelper(p_277724_, BlockEntityType.SIGN, SignBlockEntity::tick);
+        return null; // CraftBukkit - remove unnecessary sign ticking
     }
 }
