--- a/net/minecraft/world/level/block/SignBlock.java
+++ b/net/minecraft/world/level/block/SignBlock.java
@@ -130,7 +_,7 @@
                 } else if (executedClickCommand) {
                     return InteractionResult.SUCCESS_SERVER;
                 } else if (!this.otherPlayerIsEditingSign(player, sign) && player.mayBuild() && this.hasEditableText(player, sign, isFrontText)) {
-                    this.openTextEdit(player, sign, isFrontText);
+                    this.openTextEdit(player, sign, isFrontText, org.bukkit.event.player.PlayerSignOpenEvent.Cause.INTERACT); // CraftBukkit
                     return InteractionResult.SUCCESS_SERVER;
                 } else {
                     return InteractionResult.PASS;
@@ -170,6 +_,15 @@
     }
 
     public void openTextEdit(Player player, SignBlockEntity sign, boolean isFrontText) {
+        // Craftbukkit start
+        openTextEdit(player, sign, isFrontText, org.bukkit.event.player.PlayerSignOpenEvent.Cause.UNKNOWN);
+    }
+
+    public void openTextEdit(Player player, SignBlockEntity sign, boolean isFrontText, org.bukkit.event.player.PlayerSignOpenEvent.Cause cause) {
+        if (!org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerSignOpenEvent(player, sign, isFrontText, cause)) {
+            return;
+        }
+        // Craftbukkit end
         sign.setAllowedPlayerEditor(player.getUUID());
         player.openTextEdit(sign, isFrontText);
     }
