--- a/net/minecraft/world/level/block/SkullBlock.java
+++ b/net/minecraft/world/level/block/SkullBlock.java
@@ -22,7 +_,6 @@
     public static final IntegerProperty ROTATION = BlockStateProperties.ROTATION_16;
     private static final VoxelShape SHAPE = Block.column(8.0, 0.0, 8.0);
     private static final VoxelShape SHAPE_PIGLIN = Block.column(10.0, 0.0, 8.0);
-    private static final VoxelShape SHAPE_DRAGON_OUTLINE = Block.column(8.0, 0.0, 8.5);
 
     public SkullBlock(SkullBlock.Type type, BlockBehaviour.Properties properties) {
         super(type, properties);
@@ -31,11 +_,6 @@
 
     @Override
     protected VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
-        return this.getType() == SkullBlock.Types.DRAGON ? SHAPE_DRAGON_OUTLINE : this.getCollisionShape(state, level, pos, context);
-    }
-
-    @Override
-    protected VoxelShape getCollisionShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
         return this.getType() == SkullBlock.Types.PIGLIN ? SHAPE_PIGLIN : SHAPE;
     }
 
