--- a/net/minecraft/world/level/block/SmallDripleafBlock.java
+++ b/net/minecraft/world/level/block/SmallDripleafBlock.java
@@ -63,7 +_,7 @@
             BlockState blockState = DoublePlantBlock.copyWaterloggedFrom(
                 level, abovePos, this.defaultBlockState().setValue(HALF, DoubleBlockHalf.UPPER).setValue(FACING, state.getValue(FACING))
             );
-            level.setBlockAndUpdate(abovePos, blockState);
+            level.setBlock(abovePos, blockState, 3);
         }
     }
 
@@ -80,6 +_,8 @@
 
         BlockPos belowPos = pos.below();
         BlockState belowState = level.getBlockState(belowPos);
+        var soilDecision = belowState.canSustainPlant(level, belowPos, Direction.UP, state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         return this.mayPlaceOn(belowState, level, belowPos);
     }
 
