--- a/net/minecraft/world/level/block/SmokerBlock.java
+++ b/net/minecraft/world/level/block/SmokerBlock.java
@@ -44,8 +_,7 @@
     @Override
     protected void openContainer(Level p_56443_, BlockPos p_56444_, Player p_56445_) {
         BlockEntity blockentity = p_56443_.getBlockEntity(p_56444_);
-        if (blockentity instanceof SmokerBlockEntity) {
-            p_56445_.openMenu((MenuProvider)blockentity);
+        if (blockentity instanceof SmokerBlockEntity && p_56445_.openMenu((MenuProvider)blockentity).isPresent()) { // Paper - Fix InventoryOpenEvent cancellation
             p_56445_.awardStat(Stats.INTERACT_WITH_SMOKER);
         }
     }
