--- a/net/minecraft/world/level/block/SoundType.java
+++ b/net/minecraft/world/level/block/SoundType.java
@@ -847,15 +_,6 @@
     public static final SoundType CINNABAR = new SoundType(
         1.0F, 1.0F, SoundEvents.CINNABAR_BREAK, SoundEvents.CINNABAR_STEP, SoundEvents.CINNABAR_PLACE, SoundEvents.CINNABAR_HIT, SoundEvents.CINNABAR_FALL
     );
-    public static final SoundType SHELF_MUSHROOM = new SoundType(
-        1.0F,
-        1.0F,
-        SoundEvents.SHELF_MUSHROOM_BREAK,
-        SoundEvents.SHELF_MUSHROOM_STEP,
-        SoundEvents.SHELF_MUSHROOM_PLACE,
-        SoundEvents.EMPTY,
-        SoundEvents.SHELF_MUSHROOM_FALL
-    );
     public final float volume;
     public final float pitch;
     public final SoundEvent breakSound;
@@ -864,6 +_,7 @@
     public final SoundEvent hitSound;
     private final SoundEvent fallSound;
 
+    @Deprecated // Forge: Use {@link net.neoforged.neoforge.common.util.DeferredSoundType} instead for suppliers
     public SoundType(float volume, float pitch, SoundEvent breakSound, SoundEvent stepSound, SoundEvent placeSound, SoundEvent hitSound, SoundEvent fallSound) {
         this.volume = volume;
         this.pitch = pitch;
