--- a/net/minecraft/world/level/block/SoundType.java
+++ b/net/minecraft/world/level/block/SoundType.java
@@ -847,30 +_,6 @@
     public static final SoundType CINNABAR = new SoundType(
         1.0F, 1.0F, SoundEvents.CINNABAR_BREAK, SoundEvents.CINNABAR_STEP, SoundEvents.CINNABAR_PLACE, SoundEvents.CINNABAR_HIT, SoundEvents.CINNABAR_FALL
     );
-    public static final SoundType SHELF_MUSHROOM = new SoundType(
-        1.0F,
-        1.0F,
-        SoundEvents.SHELF_MUSHROOM_BREAK,
-        SoundEvents.SHELF_MUSHROOM_STEP,
-        SoundEvents.SHELF_MUSHROOM_PLACE,
-        SoundEvents.EMPTY,
-        SoundEvents.SHELF_MUSHROOM_FALL
-    );
-    public static final SoundType POPLAR_LEAVES = new SoundType(
-        1.0F,
-        1.0F,
-        SoundEvents.POPLAR_LEAVES_BREAK,
-        SoundEvents.POPLAR_LEAVES_STEP,
-        SoundEvents.POPLAR_LEAVES_PLACE,
-        SoundEvents.POPLAR_LEAVES_HIT,
-        SoundEvents.POPLAR_LEAVES_FALL
-    );
-    public static final SoundType STRAW_BED = new SoundType(
-        1.0F, 1.0F, SoundEvents.STRAW_BED_BREAK, SoundEvents.STRAW_BED_STEP, SoundEvents.STRAW_BED_PLACE, SoundEvents.STRAW_BED_HIT, SoundEvents.STRAW_BED_FALL
-    );
-    public static final SoundType RED_SHRUB = new SoundType(
-        1.0F, 1.0F, SoundEvents.RED_SHRUB_BREAK, SoundEvents.EMPTY, SoundEvents.RED_SHRUB_PLACE, SoundEvents.EMPTY, SoundEvents.EMPTY
-    );
     public final float volume;
     public final float pitch;
     public final SoundEvent breakSound;
@@ -879,6 +_,7 @@
     public final SoundEvent hitSound;
     private final SoundEvent fallSound;
 
+    @Deprecated // Forge: Use {@link net.neoforged.neoforge.common.util.DeferredSoundType} instead for suppliers
     public SoundType(float volume, float pitch, SoundEvent breakSound, SoundEvent stepSound, SoundEvent placeSound, SoundEvent hitSound, SoundEvent fallSound) {
         this.volume = volume;
         this.pitch = pitch;
