--- a/net/minecraft/world/level/block/SoundType.java
+++ b/net/minecraft/world/level/block/SoundType.java
@@ -873,12 +_,13 @@
     );
     public final float volume;
     public final float pitch;
-    private final SoundEvent breakSound;
+    public final SoundEvent breakSound;
     private final SoundEvent stepSound;
     private final SoundEvent placeSound;
-    private final SoundEvent hitSound;
+    public final SoundEvent hitSound;
     private final SoundEvent fallSound;
 
+    @Deprecated // Forge: Use {@link net.neoforged.neoforge.common.util.DeferredSoundType} instead for suppliers
     public SoundType(float volume, float pitch, SoundEvent breakSound, SoundEvent stepSound, SoundEvent placeSound, SoundEvent hitSound, SoundEvent fallSound) {
         this.volume = volume;
         this.pitch = pitch;
