--- a/net/minecraft/world/level/block/SpongeBlock.java
+++ b/net/minecraft/world/level/block/SpongeBlock.java
@@ -1,7 +_,6 @@
 package net.minecraft.world.level.block;
 
 import com.mojang.serialization.MapCodec;
-import java.util.function.Consumer;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.sounds.SoundEvents;
@@ -12,6 +_,8 @@
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.craftbukkit.util.BlockStateListPopulator;
 
 public class SpongeBlock extends Block {
     public static final MapCodec<SpongeBlock> CODEC = simpleCodec(SpongeBlock::new);
@@ -49,7 +_,10 @@
     }
 
     private boolean removeWaterBreadthFirstSearch(Level p_56808_, BlockPos p_56809_) {
-        return BlockPos.breadthFirstTraversal(
+        BlockState spongeState = p_56808_.getBlockState(p_56809_);
+        // CraftBukkit start - Use BlockStateListPopulator
+        BlockStateListPopulator blockList = new BlockStateListPopulator(p_56808_);
+        BlockPos.breadthFirstTraversal(
                 p_56809_,
                 6,
                 65,
@@ -62,18 +_,20 @@
                     if (p_294069_.equals(p_56809_)) {
                         return true;
                     } else {
-                        BlockState blockstate = p_56808_.getBlockState(p_294069_);
-                        FluidState fluidstate = p_56808_.getFluidState(p_294069_);
-                        if (!fluidstate.is(FluidTags.WATER)) {
+                    // CraftBukkit - Use blockList instead of direct world access
+                    BlockState blockstate = blockList.getBlockState(p_294069_);
+                    FluidState fluidstate = blockList.getFluidState(p_294069_);
+
+                        if (!spongeState.canBeHydrated(p_56808_, p_56809_, fluidstate, p_294069_)) {
                             return false;
                         } else {
                             if (blockstate.getBlock() instanceof BucketPickup bucketpickup
-                                && !bucketpickup.pickupBlock(null, p_56808_, p_294069_, blockstate).isEmpty()) {
+                            && !bucketpickup.pickupBlock(null, blockList, p_294069_, blockstate).isEmpty()) {
                                 return true;
                             }
 
                             if (blockstate.getBlock() instanceof LiquidBlock) {
-                                p_56808_.setBlock(p_294069_, Blocks.AIR.defaultBlockState(), 3);
+                            blockList.setBlock(p_294069_, Blocks.AIR.defaultBlockState(), 3);
                             } else {
                                 if (!blockstate.is(Blocks.KELP)
                                     && !blockstate.is(Blocks.KELP_PLANT)
@@ -82,16 +_,54 @@
                                     return false;
                                 }
 
-                                BlockEntity blockentity = blockstate.hasBlockEntity() ? p_56808_.getBlockEntity(p_294069_) : null;
-                                dropResources(blockstate, p_56808_, p_294069_, blockentity);
-                                p_56808_.setBlock(p_294069_, Blocks.AIR.defaultBlockState(), 3);
+                            // Don't drop items in populators
+                            blockList.setBlock(p_294069_, Blocks.AIR.defaultBlockState(), 3);
+                        }
+
+                        return true;
+                    }
+                }
+            }
+        );
+
+        java.util.List<CraftBlockState> blocks = blockList.getList(); // Is a clone
+        if (!blocks.isEmpty()) {
+            final org.bukkit.block.Block bblock = p_56808_.getWorld().getBlockAt(p_56809_.getX(), p_56809_.getY(), p_56809_.getZ());
+
+            org.bukkit.event.block.SpongeAbsorbEvent event = new org.bukkit.event.block.SpongeAbsorbEvent(bblock, (java.util.List<org.bukkit.block.BlockState>) (java.util.List) blocks);
+            p_56808_.getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return false;
+            }
+
+            for (CraftBlockState block : blocks) {
+                BlockPos blockPos = block.getPosition();
+                BlockState blockState = p_56808_.getBlockState(blockPos);
+                FluidState fluidState = p_56808_.getFluidState(blockPos);
+
+                if (fluidState.is(FluidTags.WATER)) {
+                    if (blockState.getBlock() instanceof BucketPickup bucketPickup
+                            && !bucketPickup.pickupBlock(null, blockList, blockPos, blockState).isEmpty()) {
+                        // NOP
+                    } else if (blockState.getBlock() instanceof LiquidBlock) {
+                        // NOP
+                    } else if (blockState.is(Blocks.KELP) || blockState.is(Blocks.KELP_PLANT)
+                            || blockState.is(Blocks.SEAGRASS) || blockState.is(Blocks.TALL_SEAGRASS)) {
+                        BlockEntity blockEntity = blockState.hasBlockEntity() ? p_56808_.getBlockEntity(blockPos) : null;
+                        // Paper start - Fix SpongeAbsortEvent handling
+                        if (block.getHandle().isAir()) {
+                        dropResources(blockState, p_56808_, blockPos, blockEntity);
+                        }
+                        // Paper end - Fix SpongeAbsortEvent handling
+                    }
+                }
+                p_56808_.setBlock(blockPos, block.getHandle(), block.getFlag());
                             }
 
                             return true;
                         }
-                    }
-                }
-            )
-            > 1;
+        return false;
+        // CraftBukkit end
     }
 }
