--- a/net/minecraft/world/level/block/StrawBedBlock.java
+++ b/net/minecraft/world/level/block/StrawBedBlock.java
@@ -1,7 +_,6 @@
 package net.minecraft.world.level.block;
 
 import java.util.Map;
-import java.util.OptionalDouble;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.resources.Identifier;
@@ -60,15 +_,12 @@
     }
 
     @Override
-    public OptionalDouble getSleepHeight(BlockState state, Level level, BlockPos pos) {
+    public double getSleepHeight(BlockState state, Level level, BlockPos pos) {
         BlockPos layingOnPos;
         BlockState layingOnState;
         if (state.getValue(BedBlock.PART) == BedPart.HEAD) {
             layingOnPos = pos.relative(getConnectedDirection(state));
             layingOnState = level.getBlockState(layingOnPos);
-            if (!layingOnState.is(this) || layingOnState.getValue(BedBlock.PART) != BedPart.FOOT) {
-                return OptionalDouble.empty();
-            }
         } else {
             layingOnPos = pos;
             layingOnState = state;
