--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -55,14 +_,18 @@
                 height++;
             }
 
-            if (height < 3) {
+            if (height < level.paperConfig().maxGrowthHeight.reeds) { // Paper - Configurable cactus/bamboo/reed growth height
+                int modifier = level.spigotConfig.caneModifier; // Spigot - SPIGOT-7159: Better modifier resolution
                 int age = state.getValue(AGE);
-                if (age == 15) {
-                    level.setBlockAndUpdate(pos.above(), this.defaultBlockState());
+                if (net.neoforged.neoforge.common.CommonHooks.canCropGrow(level, pos, state, true)) {
+                if (age >= 15 || (modifier != 100 && random.nextFloat() < (modifier / (100.0F * 16)))) { // Spigot - SPIGOT-7159: Better modifier resolution
+                    org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos.above(), this.defaultBlockState(), Block.UPDATE_ALL); // CraftBukkit
+                    net.neoforged.neoforge.common.CommonHooks.fireCropGrowPost(level, pos.above(), this.defaultBlockState());
                     level.setBlock(pos, state.setValue(AGE, 0), 260);
-                } else {
+                } else if (modifier == 100 || random.nextFloat() < (modifier / (100.0F * 16))) { // Spigot - SPIGOT-7159: Better modifier resolution
                     level.setBlock(pos, state.setValue(AGE, age + 1), 260);
                 }
+                }
             }
         }
     }
@@ -92,13 +_,16 @@
             return true;
         }
 
+        var soilDecision = stateBelow.canSustainPlant(level, pos.below(), Direction.UP, state);
+        if (!soilDecision.isDefault()) return soilDecision.isTrue();
         if (stateBelow.is(BlockTags.SUPPORTS_SUGAR_CANE)) {
             BlockPos below = pos.below();
 
             for (Direction direction : Direction.Plane.HORIZONTAL) {
                 BlockState blockState = level.getBlockState(below.relative(direction));
                 FluidState fluidState = level.getFluidState(below.relative(direction));
-                if (fluidState.is(FluidTags.SUPPORTS_SUGAR_CANE_ADJACENTLY) || blockState.is(BlockTags.SUPPORTS_SUGAR_CANE_ADJACENTLY)) {
+                // Neo: Allow the fluid state to dynamically decide whether it hydrates here if the tag-based checks fail
+                if (fluidState.is(FluidTags.SUPPORTS_SUGAR_CANE_ADJACENTLY) || blockState.is(BlockTags.SUPPORTS_SUGAR_CANE_ADJACENTLY) || state.canBeHydrated(level, pos, fluidState, below.relative(direction))) {
                     return true;
                 }
             }
