--- a/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -35,6 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.BlockRedstoneEvent;
 import org.jspecify.annotations.Nullable;
 
 public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleWaterloggedBlock {
@@ -127,6 +_,19 @@
         if (!level.isClientSide()) {
             boolean signal = level.hasNeighborSignal(pos);
             if (signal != state.getValue(POWERED)) {
+                // CraftBukkit start
+                org.bukkit.World bworld = level.getWorld();
+                org.bukkit.block.Block bblock = bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+
+                int power = bblock.getBlockPower();
+                int oldPower = state.getValue(OPEN) ? 15 : 0;
+
+                if (oldPower == 0 ^ power == 0 || block.defaultBlockState().isSignalSource()) {
+                    BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bblock, oldPower, power);
+                    level.getCraftServer().getPluginManager().callEvent(eventRedstone);
+                    signal = eventRedstone.getNewCurrent() > 0;
+                }
+                // CraftBukkit end
                 if (state.getValue(OPEN) != signal) {
                     state = state.setValue(OPEN, signal);
                     this.playSound(null, level, pos, signal);
@@ -185,6 +_,17 @@
         }
 
         return super.updateShape(state, level, ticks, pos, directionToNeighbour, neighbourPos, neighbourState, random);
+    }
+
+    // Neo: Allows Trapdoors to be climbable if any ladder, even modded ladders, is below Trapdoor
+    @Override
+    public boolean isLadder(BlockState state, net.minecraft.world.level.LevelReader world, BlockPos pos, net.minecraft.world.entity.LivingEntity entity) {
+        if (state.getValue(OPEN)) {
+            BlockPos downPos = pos.below();
+            BlockState down = world.getBlockState(downPos);
+            return down.getBlock().makesOpenTrapdoorAboveClimbable(down, world, downPos, state);
+        }
+        return false;
     }
 
     protected BlockSetType getType() {
