--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -26,6 +_,7 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.v1_21_R1.block.CraftBlock;
 
 public class TripWireHookBlock extends Block {
     public static final MapCodec<TripWireHookBlock> CODEC = simpleCodec(TripWireHookBlock::new);
@@ -176,8 +_,8 @@
                     BlockPos blockpos2 = p_57687_.relative(direction, k);
                     BlockState blockstate2 = ablockstate[k];
                     if (blockstate2 != null) {
+                    if (!p_57686_.getBlockState(blockpos2).isAir()) { // FORGE: fix MC-129055
                         p_57686_.setBlock(blockpos2, blockstate2.trySetValue(ATTACHED, Boolean.valueOf(flag2)), 3);
-                        if (!p_57686_.getBlockState(blockpos2).isAir()) {
                         }
                     }
                 }
@@ -201,6 +_,15 @@
             p_222603_.playSound(null, p_222604_, SoundEvents.TRIPWIRE_ATTACH, SoundSource.BLOCKS, 0.4F, 0.7F);
             p_222603_.gameEvent(null, GameEvent.BLOCK_ATTACH, p_222604_);
         } else if (!p_222605_ && p_222607_) {
+            // CraftBukkit start
+            org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(CraftBlock.at(p_222603_, p_222604_), 15, 0);
+            p_222603_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+            if (eventRedstone.getNewCurrent() > 0) {
+                return;
+            }
+            // CraftBukkit end
+
             p_222603_.playSound(null, p_222604_, SoundEvents.TRIPWIRE_DETACH, SoundSource.BLOCKS, 0.4F, 1.2F / (p_222603_.random.nextFloat() * 0.2F + 0.9F));
             p_222603_.gameEvent(null, GameEvent.BLOCK_DETACH, p_222604_);
         }
