--- a/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -26,6 +_,8 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class TripWireHookBlock extends Block {
     public static final MapCodec<TripWireHookBlock> CODEC = simpleCodec(TripWireHookBlock::new);
@@ -163,9 +_,19 @@
                 emitState(p_57686_, blockpos1, flag2, flag3, flag, flag1);
             }
 
+            // CraftBukkit start
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(CraftBlock.at(p_57686_, p_57687_), 15, 0);
+            p_57686_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+            if (eventRedstone.getNewCurrent() > 0) {
+                return;
+            }
+            // CraftBukkit end
+
             emitState(p_57686_, p_57687_, flag2, flag3, flag, flag1);
             if (!p_57689_) {
-                p_57686_.setBlock(p_57687_, blockstate1.setValue(FACING, direction), 3);
+                if (p_57686_.getBlockState(p_57687_).getBlock() == Blocks.TRIPWIRE_HOOK) // Paper - Validate tripwire hook placement before update
+                    p_57686_.setBlock(p_57687_, blockstate1.setValue(FACING, direction), 3);
                 if (p_57690_) {
                     notifyNeighbors(block, p_57686_, p_57687_, direction);
                 }
@@ -176,8 +_,8 @@
                     BlockPos blockpos2 = p_57687_.relative(direction, k);
                     BlockState blockstate2 = ablockstate[k];
                     if (blockstate2 != null) {
-                        p_57686_.setBlock(blockpos2, blockstate2.trySetValue(ATTACHED, Boolean.valueOf(flag2)), 3);
-                        if (!p_57686_.getBlockState(blockpos2).isAir()) {
+                        if (p_57686_.getBlockState(blockpos2).is(Blocks.TRIPWIRE) || io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowTripwireDisarmingExploits) { // Paper - Fix tripwire disarming not working as intended
+                            p_57686_.setBlock(blockpos2, blockstate2.trySetValue(ATTACHED, Boolean.valueOf(flag2)), 3);
                         }
                     }
                 }
@@ -201,6 +_,15 @@
             p_222603_.playSound(null, p_222604_, SoundEvents.TRIPWIRE_ATTACH, SoundSource.BLOCKS, 0.4F, 0.7F);
             p_222603_.gameEvent(null, GameEvent.BLOCK_ATTACH, p_222604_);
         } else if (!p_222605_ && p_222607_) {
+            // CraftBukkit start
+            org.bukkit.event.block.BlockRedstoneEvent eventRedstone = new org.bukkit.event.block.BlockRedstoneEvent(CraftBlock.at(p_222603_, p_222604_), 15, 0);
+            p_222603_.getCraftServer().getPluginManager().callEvent(eventRedstone);
+
+            if (eventRedstone.getNewCurrent() > 0) {
+                return;
+            }
+            // CraftBukkit end
+
             p_222603_.playSound(null, p_222604_, SoundEvents.TRIPWIRE_DETACH, SoundSource.BLOCKS, 0.4F, 1.2F / (p_222603_.random.nextFloat() * 0.2F + 0.9F));
             p_222603_.gameEvent(null, GameEvent.BLOCK_DETACH, p_222604_);
         }
