--- a/net/minecraft/world/level/block/WallHangingSignBlock.java
+++ b/net/minecraft/world/level/block/WallHangingSignBlock.java
@@ -18,7 +_,6 @@
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.entity.BlockEntityTicker;
 import net.minecraft.world.level.block.entity.BlockEntityType;
-import net.minecraft.world.level.block.entity.BlockEntityTypes;
 import net.minecraft.world.level.block.entity.HangingSignBlockEntity;
 import net.minecraft.world.level.block.entity.SignBlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
@@ -60,7 +_,7 @@
     private boolean shouldTryToChainAnotherHangingSign(
         BlockState state, Player player, BlockHitResult hitResult, SignBlockEntity signEntity, ItemStack itemStack
     ) {
-        return !signEntity.canExecuteClickCommands(signEntity.getSlotPlayerIsFacing(player), player)
+        return !signEntity.canExecuteClickCommands(signEntity.isFacingFrontText(player), player)
             && itemStack.getItem() instanceof HangingSignItem
             && !this.isHittingEditableSide(hitResult, state);
     }
@@ -166,7 +_,7 @@
 
     @Override
     public <T extends BlockEntity> @Nullable BlockEntityTicker<T> getTicker(Level level, BlockState blockState, BlockEntityType<T> type) {
-        return createTickerHelper(type, BlockEntityTypes.HANGING_SIGN, SignBlockEntity::tick);
+        return null; // CraftBukkit - remove unnecessary sign ticking
     }
 
     @Override
