--- a/net/minecraft/world/level/block/WitherSkullBlock.java
+++ b/net/minecraft/world/level/block/WitherSkullBlock.java
@@ -20,6 +_,7 @@
 import net.minecraft.world.level.block.state.pattern.BlockPattern;
 import net.minecraft.world.level.block.state.pattern.BlockPatternBuilder;
 import net.minecraft.world.level.block.state.predicate.BlockStatePredicate;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.jspecify.annotations.Nullable;
 
 public class WitherSkullBlock extends SkullBlock {
@@ -36,7 +_,7 @@
     }
 
     public static void checkSpawn(Level level, BlockPos pos) {
-        if (level.getBlockEntity(pos) instanceof SkullBlockEntity placedSkull) {
+        if (!level.captureBlockStates && level.getBlockEntity(pos) instanceof SkullBlockEntity placedSkull) { // CraftBukkit
             checkSpawn(level, pos, placedSkull);
         }
     }
@@ -50,7 +_,7 @@
                 if (match != null) {
                     WitherBoss witherBoss = EntityTypes.WITHER.create(level, EntitySpawnReason.TRIGGERED);
                     if (witherBoss != null) {
-                        CarvedPumpkinBlock.clearPatternBlocks(level, match);
+                        // CarvedPumpkinBlock.clearPatternBlocks(level, match); // CraftBukkit - move down
                         BlockPos spawnPos = match.getBlock(1, 2, 0).getPos();
                         witherBoss.snapTo(
                             spawnPos.getX() + 0.5,
@@ -61,12 +_,19 @@
                         );
                         witherBoss.yBodyRot = match.getForwards().getAxis() == Direction.Axis.X ? 0.0F : 90.0F;
                         witherBoss.makeInvulnerable();
+                        witherBoss.setSummoner(blockState.getBlock().placer == null ? null : blockState.getBlock().placer.getUUID()); // Purpur - Summoner API
+                        // CraftBukkit start
+                        if (!level.addFreshEntity(witherBoss, CreatureSpawnEvent.SpawnReason.BUILD_WITHER)) {
+                            return;
+                        }
+                        CarvedPumpkinBlock.clearPatternBlocks(level, match); // CraftBukkit - from above
+                        // CraftBukkit end
 
                         for (ServerPlayer player : level.getEntitiesOfClass(ServerPlayer.class, witherBoss.getBoundingBox().inflate(50.0))) {
                             CriteriaTriggers.SUMMONED_ENTITY.trigger(player, witherBoss);
                         }
 
-                        level.addFreshEntity(witherBoss);
+                        // level.addFreshEntity(witherBoss); // CraftBukkit - moved up
                         CarvedPumpkinBlock.updatePatternBlocks(level, match);
                     }
                 }
