--- a/net/minecraft/world/level/block/WitherRoseBlock.java
+++ b/net/minecraft/world/level/block/WitherRoseBlock.java
@@ -69,7 +_,7 @@
             && level.getDifficulty() != Difficulty.PEACEFUL
             && entity instanceof LivingEntity livingEntity
             && !livingEntity.isInvulnerableTo(serverLevel, level.damageSources().wither())) {
-            livingEntity.addEffect(this.getBeeInteractionEffect());
+            livingEntity.addEffect(this.getBeeInteractionEffect(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.WITHER_ROSE); // CraftBukkit
         }
     }
 
