--- a/net/minecraft/world/level/saveddata/maps/MapDecorationType.java
+++ b/net/minecraft/world/level/saveddata/maps/MapDecorationType.java
@@ -9,9 +_,14 @@
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.resources.Identifier;
 
-public record MapDecorationType(Identifier assetId, boolean showOnItemFrame, boolean trackCount) {
+public record MapDecorationType(Identifier assetId, boolean showOnItemFrame, int mapColor, boolean explorationMapElement, boolean trackCount) {
+    public static final int NO_MAP_COLOR = -1;
     public static final Codec<Holder<MapDecorationType>> CODEC = BuiltInRegistries.MAP_DECORATION_TYPE.holderByNameCodec();
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<MapDecorationType>> STREAM_CODEC = ByteBufCodecs.holderRegistry(
         Registries.MAP_DECORATION_TYPE
     );
+
+    public boolean hasMapColor() {
+        return this.mapColor != -1;
+    }
 }
