--- a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -5,12 +_,14 @@
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Dynamic;
 import io.netty.buffer.ByteBuf;
+import io.papermc.paper.adventure.PaperAdventure;
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
+import java.util.UUID;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -41,6 +_,12 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.level.saveddata.SavedData;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.map.CraftMapView;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.map.CraftMapCursor;
+import org.bukkit.craftbukkit.map.RenderData;
 import org.slf4j.Logger;
 
 public class MapItemSavedData extends SavedData {
@@ -65,6 +_,14 @@
     private final Map<String, MapFrame> frameMarkers = Maps.newHashMap();
     private int trackedDecorationCount;
 
+    // CraftBukkit start
+    public final CraftMapView mapView;
+    private CraftServer server;
+    public UUID uniqueId = null;
+    public MapId id;
+    // CraftBukkit end
+    public boolean isExplorerMap; // Purpur
+
     public static SavedData.Factory<MapItemSavedData> factory() {
         return new SavedData.Factory<>(() -> {
             throw new IllegalStateException("Should never create an empty map saved data");
@@ -82,6 +_,10 @@
         this.unlimitedTracking = p_164772_;
         this.locked = p_164773_;
         this.setDirty();
+        // CraftBukkit start
+        mapView = new CraftMapView(this);
+        server = (CraftServer) org.bukkit.Bukkit.getServer();
+        // CraftBukkit end
     }
 
     public static MapItemSavedData createFresh(
@@ -100,9 +_,29 @@
     }
 
     public static MapItemSavedData load(CompoundTag p_164808_, HolderLookup.Provider p_324560_) {
+        // CraftBukkit start
         ResourceKey<Level> resourcekey = DimensionType.parseLegacy(new Dynamic<>(NbtOps.INSTANCE, p_164808_.get("dimension")))
             .resultOrPartial(LOGGER::error)
-            .orElseThrow(() -> new IllegalArgumentException("Invalid map dimension: " + p_164808_.get("dimension")));
+            .orElseGet(() -> {
+                long least = p_164808_.getLong("UUIDLeast");
+                long most = p_164808_.getLong("UUIDMost");
+
+                if (least != 0L && most != 0L) {
+                    java.util.UUID uniqueId = new java.util.UUID(most, least);
+
+                    CraftWorld world = (CraftWorld) org.bukkit.Bukkit.getWorld(uniqueId);
+                    // Check if the stored world details are correct.
+                    if (world == null) {
+                        /* All Maps which do not have their valid world loaded are set to a dimension which hopefully won't be reached.
+                           This is to prevent them being corrupted with the wrong map data. */
+                        // PAIL: Use Vanilla exception handling for now
+                    } else {
+                        return world.getHandle().dimension();
+                    }
+                }
+                throw new IllegalArgumentException("Invalid map dimension: " + String.valueOf(p_164808_.get("dimension")));
+            });
+        // CraftBukkit end
         int i = p_164808_.getInt("xCenter");
         int j = p_164808_.getInt("zCenter");
         byte b0 = (byte)Mth.clamp(p_164808_.getByte("scale"), 0, 4);
@@ -156,10 +_,32 @@
 
     @Override
     public CompoundTag save(CompoundTag p_77956_, HolderLookup.Provider p_323858_) {
-        ResourceLocation.CODEC
-            .encodeStart(NbtOps.INSTANCE, this.dimension.location())
-            .resultOrPartial(LOGGER::error)
-            .ifPresent(p_77954_ -> p_77956_.put("dimension", p_77954_));
+        // CraftBukkit start
+        com.mojang.serialization.DataResult<net.minecraft.nbt.Tag> dataresult = ResourceLocation.CODEC.encodeStart(NbtOps.INSTANCE, this.dimension.location());
+        Logger logger = LOGGER;
+
+        Objects.requireNonNull(logger);
+        dataresult.resultOrPartial(logger::error).ifPresent((nbtbase) -> {
+            p_77956_.put("dimension", nbtbase);
+        });
+        if (true) {
+            if (this.uniqueId == null) {
+                for (org.bukkit.World world : server.getWorlds()) {
+                    CraftWorld cWorld = (CraftWorld) world;
+                    if (cWorld.getHandle().dimension() == this.dimension) {
+                        this.uniqueId = cWorld.getUID();
+                        break;
+                    }
+                }
+            }
+            /* Perform a second check to see if a matching world was found, this is a necessary
+               change incase Maps are forcefully unlinked from a World and lack a UID.*/
+            if (this.uniqueId != null) {
+                p_77956_.putLong("UUIDLeast", this.uniqueId.getLeastSignificantBits());
+                p_77956_.putLong("UUIDMost", this.uniqueId.getMostSignificantBits());
+            }
+        }
+        // CraftBukkit end
         p_77956_.putInt("xCenter", this.centerX);
         p_77956_.putInt("zCenter", this.centerZ);
         p_77956_.putByte("scale", this.scale);
@@ -246,6 +_,7 @@
             }
 
             MapFrame mapframe = new MapFrame(blockpos, itemframe.getDirection().get2DDataValue() * 90, itemframe.getId());
+            if (this.decorations.size() < p_77919_.level().paperConfig().maps.itemFrameCursorLimit) { // Paper - Limit item frame cursors on maps
             this.addDecoration(
                 MapDecorationTypes.FRAME,
                 p_77919_.level(),
@@ -256,6 +_,7 @@
                 null
             );
             this.frameMarkers.put(mapframe.getId(), mapframe);
+            } // Paper - Limit item frame cursors on maps
         }
 
         MapDecorations mapdecorations = p_77920_.getOrDefault(DataComponents.MAP_DECORATIONS, MapDecorations.EMPTY);
@@ -268,7 +_,7 @@
         }
     }
 
-    public void removeDecoration(String p_164800_) {
+    private void removeDecoration(String p_164800_) {
         MapDecoration mapdecoration = this.decorations.remove(p_164800_);
         if (mapdecoration != null && mapdecoration.type().value().trackCount()) {
             this.trackedDecorationCount--;
@@ -285,7 +_,7 @@
         }
     }
 
-    public void addDecoration(
+    private void addDecoration(
         Holder<MapDecorationType> p_335830_,
         @Nullable LevelAccessor p_77939_,
         String p_77940_,
@@ -407,7 +_,7 @@
                 return true;
             }
 
-            if (!this.isTrackedCountOverLimit(256)) {
+            if (!this.isTrackedCountOverLimit(((Level) p_77935_).paperConfig().maps.itemFrameCursorLimit)) { // Paper - Limit item frame cursors on maps
                 this.bannerMarkers.put(mapbanner.getId(), mapbanner);
                 this.addDecoration(mapbanner.getDecoration(), p_77935_, mapbanner.getId(), d0, d1, 180.0, mapbanner.name().orElse(null));
                 return true;
@@ -506,7 +_,13 @@
             this.player = p_77970_;
         }
 
+        private RenderData render;
+        private byte[] buffer;
         private MapItemSavedData.MapPatch createPatch() {
+            // CraftBukkit start
+            render = MapItemSavedData.this.mapView.render((CraftPlayer) this.player.getBukkitEntity());
+            buffer = render.buffer;
+
             int i = this.minDirtyX;
             int j = this.minDirtyY;
             int k = this.maxDirtyX + 1 - this.minDirtyX;
@@ -515,15 +_,18 @@
 
             for (int i1 = 0; i1 < k; i1++) {
                 for (int j1 = 0; j1 < l; j1++) {
-                    abyte[i1 + j1 * k] = MapItemSavedData.this.colors[i + i1 + (j + j1) * 128];
+                    abyte[i1 + j1 * k] = buffer[i + i1 + (j + j1) * 128];
                 }
             }
+            // CraftBukkit end
 
             return new MapItemSavedData.MapPatch(i, j, k, l, abyte);
         }
 
         @Nullable
         Packet<?> nextUpdatePacket(MapId p_324558_) {
+            RenderData render = MapItemSavedData.this.mapView.render((CraftPlayer) this.player.getBukkitEntity()); // CraftBukkit
+
             MapItemSavedData.MapPatch mapitemsaveddata$mappatch;
             if (this.dirtyData) {
                 this.dirtyData = false;
@@ -533,9 +_,25 @@
             }
 
             Collection<MapDecoration> collection;
-            if (this.dirtyDecorations && this.tick++ % 5 == 0) {
+            // CraftBukkit start
+            if ((true || this.dirtyDecorations) && this.tick++ % 5 == 0) { // custom maps don't update this yet
                 this.dirtyDecorations = false;
-                collection = MapItemSavedData.this.decorations.values();
+
+                java.util.Collection<MapDecoration> icons = new java.util.ArrayList<MapDecoration>();
+
+                for (org.bukkit.map.MapCursor cursor : render.cursors) {
+                    if (cursor.isVisible()) {
+                        icons.add(new MapDecoration(
+                            CraftMapCursor.CraftType.bukkitToMinecraftHolder(cursor.getType()),
+                            cursor.getX(),
+                            cursor.getY(),
+                            cursor.getDirection(),
+                            Optional.ofNullable(PaperAdventure.asVanilla(cursor.caption())
+                        )));
+                    }
+                }
+                collection = icons;
+                // CraftBukkit end
             } else {
                 collection = null;
             }
