--- a/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
+++ b/net/minecraft/world/level/saveddata/maps/MapItemSavedData.java
@@ -268,7 +_,7 @@
         }
     }
 
-    public void removeDecoration(String p_164800_) {
+    private void removeDecoration(String p_164800_) {
         MapDecoration mapdecoration = this.decorations.remove(p_164800_);
         if (mapdecoration != null && mapdecoration.type().value().trackCount()) {
             this.trackedDecorationCount--;
@@ -285,7 +_,7 @@
         }
     }
 
-    public void addDecoration(
+    private void addDecoration(
         Holder<MapDecorationType> p_335830_,
         @Nullable LevelAccessor p_77939_,
         String p_77940_,
