--- a/net/minecraft/server/network/LegacyQueryHandler.java
+++ b/net/minecraft/server/network/LegacyQueryHandler.java
@@ -6,6 +_,8 @@
 import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelInboundHandlerAdapter;
+import com.destroystokyo.paper.event.server.PaperServerListPingEvent;
+import com.destroystokyo.paper.network.PaperLegacyStatusClient;
 import java.net.SocketAddress;
 import java.util.Locale;
 import net.minecraft.server.ServerInfo;
@@ -33,9 +_,21 @@
 
                 SocketAddress socketaddress = p_9686_.channel().remoteAddress();
                 int i = bytebuf.readableBytes();
+                String s = null; // Paper
+                PaperServerListPingEvent event; // Paper
+
                 if (i == 0) {
-                    LOGGER.debug("Ping: (<1.3.x) from {}", socketaddress);
-                    String s = createVersion0Response(this.server);
+                    LOGGER.debug("Ping: (<1.3.x) from {}", net.minecraft.server.MinecraftServer.getServer().logIPs() ? socketaddress: "<ip address withheld>"); // Paper - Respect logIPs option
+                    // Paper start - Call PaperServerListPingEvent and use results
+                    event = PaperLegacyStatusClient.processRequest(net.minecraft.server.MinecraftServer.getServer(), (java.net.InetSocketAddress) socketaddress, 39, null);
+                    if (event == null) {
+                        p_9686_.close();
+                        bytebuf.release();
+                        flag = false;
+                        return;
+                    }
+                    s = String.format(Locale.ROOT, "%s\u00a7%d\u00a7%d", PaperLegacyStatusClient.getUnformattedMotd(event), event.getNumPlayers(), event.getMaxPlayers());
+                    // Paper end
                     sendFlushAndClose(p_9686_, createLegacyDisconnectPacket(p_9686_.alloc(), s));
                 } else {
                     if (bytebuf.readUnsignedByte() != 1) {
@@ -49,11 +_,22 @@
 
                         LOGGER.debug("Ping: (1.6) from {}", socketaddress);
                     } else {
-                        LOGGER.debug("Ping: (1.4-1.5.x) from {}", socketaddress);
+                        LOGGER.debug("Ping: (1.4-1.5.x) from {}", net.minecraft.server.MinecraftServer.getServer().logIPs() ? socketaddress: "<ip address withheld>"); // Paper - Respect logIPs option
                     }
 
-                    String s1 = createVersion1Response(this.server);
-                    sendFlushAndClose(p_9686_, createLegacyDisconnectPacket(p_9686_.alloc(), s1));
+                    if (s == null) {
+                        // Paper start - Call PaperServerListPingEvent and use results
+                        event = PaperLegacyStatusClient.processRequest(net.minecraft.server.MinecraftServer.getServer(), (java.net.InetSocketAddress) socketaddress, 127, null); // Paper
+                        if (event == null) {
+                            p_9686_.close();
+                            bytebuf.release();
+                            flag = false;
+                            return;
+                        }
+                        s = String.format(Locale.ROOT, "\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d", new Object[] { event.getProtocolVersion(), this.server.getServerVersion(), event.getMotd(), event.getNumPlayers(), event.getMaxPlayers()}); // CraftBukkit
+                        // Paper end
+                    }
+                    sendFlushAndClose(p_9686_, createLegacyDisconnectPacket(p_9686_.alloc(), s));
                 }
 
                 bytebuf.release();
@@ -95,19 +_,19 @@
         }
     }
 
-    private static String createVersion0Response(ServerInfo p_294449_) {
-        return String.format(Locale.ROOT, "%s\u00a7%d\u00a7%d", p_294449_.getMotd(), p_294449_.getPlayerCount(), p_294449_.getMaxPlayers());
+    private static String createVersion0Response(ServerInfo p_294449_, org.bukkit.event.server.ServerListPingEvent event) {
+        return String.format(Locale.ROOT, "%s\u00a7%d\u00a7%d", event.getMotd(), event.getNumPlayers(), event.getMaxPlayers());
     }
 
-    private static String createVersion1Response(ServerInfo p_295503_) {
+    private static String createVersion1Response(ServerInfo p_295503_, org.bukkit.event.server.ServerListPingEvent event) {
         return String.format(
             Locale.ROOT,
             "\u00a71\u0000%d\u0000%s\u0000%s\u0000%d\u0000%d",
             127,
             p_295503_.getServerVersion(),
-            p_295503_.getMotd(),
-            p_295503_.getPlayerCount(),
-            p_295503_.getMaxPlayers()
+            event.getMotd(),
+            event.getNumPlayers(),
+            event.getMaxPlayers()
         );
     }
 
