--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -22,6 +_,7 @@
 import java.util.Set;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
@@ -54,6 +_,7 @@
 import net.minecraft.network.chat.LastSeenMessagesValidator;
 import net.minecraft.network.chat.MessageSignature;
 import net.minecraft.network.chat.MessageSignatureCache;
+import net.minecraft.network.chat.OutgoingChatMessage;
 import net.minecraft.network.chat.PlayerChatMessage;
 import net.minecraft.network.chat.RemoteChatSession;
 import net.minecraft.network.chat.SignableCommand;
@@ -218,6 +_,13 @@
 import net.minecraft.world.phys.shapes.BooleanOp;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.util.LazyPlayerSet;
+import org.bukkit.craftbukkit.util.Waitable;
+import org.bukkit.event.player.AsyncPlayerChatEvent;
+import org.bukkit.event.player.PlayerChatEvent;
+import org.bukkit.event.player.PlayerCommandPreprocessEvent;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -276,7 +_,7 @@
     private int clientLoadedTimeoutTimer;
 
     public ServerGamePacketListenerImpl(MinecraftServer server, Connection connection, ServerPlayer player, CommonListenerCookie cookie) {
-        super(server, connection, cookie);
+        super(server, connection, cookie, player);
         this.restartClientLoadTimerAfterRespawn();
         this.chunkSender = new PlayerChunkSender(connection.isMemoryConnection());
         this.player = player;
@@ -286,6 +_,22 @@
         this.chatMessageChain = new FutureChain(server);
     }
 
+    // CraftBukkit start - add fields and methods
+    private int lastTick = MinecraftServer.currentTick;
+    private int allowedPlayerTicks = 1;
+    private int lastDropTick = MinecraftServer.currentTick;
+    private int lastBookTick  = MinecraftServer.currentTick;
+    private int dropCount = 0;
+
+    private boolean hasMoved = false;
+    private double lastPosX = Double.MAX_VALUE;
+    private double lastPosY = Double.MAX_VALUE;
+    private double lastPosZ = Double.MAX_VALUE;
+    private float lastPitch = Float.MAX_VALUE;
+    private float lastYaw = Float.MAX_VALUE;
+    private boolean justTeleported = false;
+    // CraftBukkit end
+
     @Override
     public void tick() {
         if (this.ackBlockChangesUpTo > -1) {
@@ -492,18 +_,21 @@
 
                 if (fail && level.noCollision(vehicle, oldAABB) || this.isEntityCollidingWithAnythingNew(level, vehicle, oldAABB, targetX, targetY, targetZ)) {
                     vehicle.absSnapTo(oldX, oldY, oldZ, targetYRot, targetXRot);
+                    resyncPlayerWithVehicle(vehicle); // Neo - Resync player position on vehicle moving
                     this.send(ClientboundMoveVehiclePacket.fromEntity(vehicle));
                     vehicle.removeLatestMovementRecording();
                     return;
                 }
 
                 vehicle.absSnapTo(targetX, targetY, targetZ, targetYRot, targetXRot);
+                resyncPlayerWithVehicle(vehicle); // Neo - Resync player position on vehicle moving
                 this.player.level().getChunkSource().move(this.player);
                 Vec3 clientDeltaMovement = new Vec3(vehicle.getX() - oldX, vehicle.getY() - oldY, vehicle.getZ() - oldZ);
                 this.handlePlayerKnownMovement(clientDeltaMovement);
                 vehicle.setOnGroundWithMovement(packet.onGround(), clientDeltaMovement);
                 vehicle.doCheckFallDamage(clientDeltaMovement.x, clientDeltaMovement.y, clientDeltaMovement.z, packet.onGround());
                 this.player.checkMovementStatistics(clientDeltaMovement.x, clientDeltaMovement.y, clientDeltaMovement.z);
+                this.player.checkRidingStatistics(clientDeltaMovement.x, clientDeltaMovement.y, clientDeltaMovement.z); // Neo: check riding stats too as vanilla checks them in rideTick based on the assumption that Entity#rideTick will move the entity, which we break
                 this.clientVehicleIsFloating = oyDist >= -0.03125
                     && !vehicleRestsOnSomething
                     && !this.server.allowFlight()
@@ -517,6 +_,23 @@
         }
     }
 
+    private void resyncPlayerWithVehicle(Entity vehicle) {
+        Vec3 oldPos = this.player.position();
+        float yRot = this.player.getYRot();
+        float xRot = this.player.getXRot();
+        float yHeadRot = this.player.getYHeadRot();
+
+        vehicle.positionRider(this.player);
+
+        // preserve old rotation and store old position in xo/yo/zo
+        this.player.setYRot(yRot);
+        this.player.setXRot(xRot);
+        this.player.setYHeadRot(yHeadRot);
+        this.player.xo = oldPos.x;
+        this.player.yo = oldPos.y;
+        this.player.zo = oldPos.z;
+    }
+
     private boolean noBlocksAround(Entity entity) {
         return entity.level()
             .getBlockStates(entity.getBoundingBox().inflate(0.0625).expandTowards(0.0, -0.55, 0.0))
@@ -700,7 +_,7 @@
             if (level.isLoaded(pos)) {
                 BlockState blockState = level.getBlockState(pos);
                 boolean includeData = this.player.hasInfiniteMaterials() && packet.includeData();
-                ItemStack itemStack = blockState.getCloneItemStack(level, pos, includeData);
+                ItemStack itemStack = blockState.getCloneItemStack(pos, level, includeData, player);
                 if (!itemStack.isEmpty()) {
                     if (includeData) {
                         addBlockDataToItem(blockState, level, pos, itemStack);
@@ -808,7 +_,7 @@
 
     private <T> void setGameRuleValue(GameRules gameRules, GameRule<T> rule, String value) {
         rule.deserialize(value).result().ifPresent(parsedValue -> {
-            gameRules.set(rule, (T)parsedValue, this.server);
+            gameRules.set(rule, (T)parsedValue, this.player.level()); // CraftBukkit - per-world
             this.broadcastGameRuleChangeToOperators(rule, (T)parsedValue);
         });
     }
@@ -1157,7 +_,7 @@
                                         && !playerStandsOnSomething
                                         && !this.player.isSpectator()
                                         && !this.server.allowFlight()
-                                        && !this.player.getAbilities().mayfly
+                                        && !this.player.mayFly()
                                         && !this.player.hasEffect(MobEffects.LEVITATION)
                                         && !isFallFlying
                                         && !isAutoSpinAttack
@@ -1245,11 +_,32 @@
         return false;
     }
 
+    public void teleport(Location dest) {
+        this.teleport(dest.getX(), dest.getY(), dest.getZ(), dest.getYaw(), dest.getPitch());
+    }
+
     public void teleport(double x, double y, double z, float yRot, float xRot) {
         this.teleport(new PositionMoveRotation(new Vec3(x, y, z), Vec3.ZERO, yRot, xRot), Collections.emptySet());
     }
 
+    private void internalTeleport(double d0, double d1, double d2, float f, float f1) {
+        this.teleport(d0, d1, d2, f, f1);
+    }
+
+    public void internalTeleport(PositionMoveRotation destination, Set<Relative> relatives) {
+        this.teleport(destination, relatives);
+    }
+
     public void teleport(PositionMoveRotation destination, Set<Relative> relatives) {
+        if (Float.isNaN(destination.yRot())) {
+            destination = new PositionMoveRotation(destination.position(), destination.deltaMovement(), 0, destination.xRot());
+        }
+        if (Float.isNaN(destination.xRot())) {
+            destination = new PositionMoveRotation(destination.position(), destination.deltaMovement(), destination.yRot(), 0);
+        }
+
+        this.justTeleported = true;
+        // CraftBukkit end
         this.awaitingTeleportTime = this.tickCount;
         if (++this.awaitingTeleport == Integer.MAX_VALUE) {
             this.awaitingTeleport = 0;
@@ -1257,6 +_,13 @@
 
         this.player.teleportSetPosition(destination, relatives);
         this.awaitingPositionFromClient = this.player.position();
+        // CraftBukkit start - update last location
+        this.lastPosX = this.awaitingPositionFromClient.x;
+        this.lastPosY = this.awaitingPositionFromClient.y;
+        this.lastPosZ = this.awaitingPositionFromClient.z;
+        this.lastYaw = this.player.getYRot();
+        this.lastPitch = this.player.getXRot();
+        // CraftBukkit end
         this.send(ClientboundPlayerPositionPacket.of(this.awaitingTeleport, destination, relatives));
     }
 
@@ -1286,9 +_,10 @@
                     }
                 case SWAP_ITEM_WITH_OFFHAND:
                     if (!this.player.isSpectator()) {
-                        ItemStack swap = this.player.getItemInHand(InteractionHand.OFF_HAND);
-                        this.player.setItemInHand(InteractionHand.OFF_HAND, this.player.getItemInHand(InteractionHand.MAIN_HAND));
-                        this.player.setItemInHand(InteractionHand.MAIN_HAND, swap);
+                        var event = net.neoforged.neoforge.common.CommonHooks.onLivingSwapHandItems(this.player);
+                        if (event.isCanceled()) return;
+                        this.player.setItemInHand(InteractionHand.OFF_HAND, event.getItemSwappedToOffHand());
+                        this.player.setItemInHand(InteractionHand.MAIN_HAND, event.getItemSwappedToMainHand());
                         this.player.stopUsingItem();
                     }
 
@@ -1312,7 +_,7 @@
                 case ABORT_DESTROY_BLOCK:
                 case STOP_DESTROY_BLOCK:
                     this.player.gameMode.handleBlockBreakAction(pos, action, packet.getDirection(), this.player.level().getMaxY(), packet.getSequence());
-                    this.ackBlockChangesUpTo(packet.getSequence());
+                    this.player.connection.ackBlockChangesUpTo = packet.getSequence();
                     return;
                 default:
                     throw new IllegalArgumentException("Invalid player action");
@@ -1334,7 +_,7 @@
     public void handleUseItemOn(ServerboundUseItemOnPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
         if (this.hasClientLoaded()) {
-            this.ackBlockChangesUpTo(packet.getSequence());
+            this.player.connection.ackBlockChangesUpTo = packet.getSequence();
             ServerLevel level = this.player.level();
             InteractionHand hand = packet.getHand();
             ItemStack itemStack = this.player.getItemInHand(hand);
@@ -1504,8 +_,9 @@
                 }
 
                 CompletableFuture<FilteredText> filteredFuture = this.filterTextPacket(signedMessage.signedContent());
-                Component decorated = this.server.getChatDecorator().decorate(this.player, signedMessage.decoratedContent());
+                Component decorated = net.neoforged.neoforge.common.CommonHooks.getServerChatSubmittedDecorator().decorate(this.player, signedMessage.decoratedContent());
                 this.chatMessageChain.append(filteredFuture, filtered -> {
+                    if (decorated == null) return; // Forge: ServerChatEvent was canceled if this is null.
                     PlayerChatMessage filteredMessage = signedMessage.withUnsignedContent(decorated).filter(filtered.mask());
                     this.broadcastChatMessage(filteredMessage);
                 });
@@ -1658,6 +_,116 @@
         return false;
     }
 
+    // CraftBukkit start - add method
+    public void chat(String s, PlayerChatMessage original, boolean async) {
+        if (s.isEmpty() || this.player.getChatVisibility() == ChatVisiblity.HIDDEN) {
+            return;
+        }
+        OutgoingChatMessage outgoing = OutgoingChatMessage.create(original);
+
+        if (!async && s.startsWith("/")) {
+            this.handleCommand(s);
+        } else if (this.player.getChatVisibility() == ChatVisiblity.SYSTEM) {
+            // Do nothing, this is coming from a plugin
+        } else {
+            org.bukkit.entity.Player player = this.getCraftPlayer();
+            AsyncPlayerChatEvent event = new AsyncPlayerChatEvent(async, player, s, new LazyPlayerSet(server));
+            String originalFormat = event.getFormat(), originalMessage = event.getMessage();
+            this.cserver.getPluginManager().callEvent(event);
+
+            if (PlayerChatEvent.getHandlerList().getRegisteredListeners().length != 0) {
+                // Evil plugins still listening to deprecated event
+                final PlayerChatEvent queueEvent = new PlayerChatEvent(player, event.getMessage(), event.getFormat(), event.getRecipients());
+                queueEvent.setCancelled(event.isCancelled());
+                Waitable waitable = new Waitable() {
+                    @Override
+                    protected Object evaluate() {
+                        org.bukkit.Bukkit.getPluginManager().callEvent(queueEvent);
+
+                        if (queueEvent.isCancelled()) {
+                            return null;
+                        }
+
+                        String message = String.format(queueEvent.getFormat(), queueEvent.getPlayer().getDisplayName(), queueEvent.getMessage());
+                        if (((LazyPlayerSet) queueEvent.getRecipients()).isLazy()) {
+                            if (originalFormat.equals(queueEvent.getFormat()) && originalMessage.equals(queueEvent.getMessage()) && queueEvent.getPlayer().getName().equalsIgnoreCase(queueEvent.getPlayer().getDisplayName())) {
+                                ServerGamePacketListenerImpl.this.server.getPlayerList().broadcastChatMessage(original, ServerGamePacketListenerImpl.this.player, ChatType.bind(ChatType.CHAT, (Entity) ServerGamePacketListenerImpl.this.player));
+                                return null;
+                            }
+
+                            for (ServerPlayer recipient : server.getPlayerList().players) {
+                                recipient.getBukkitEntity().sendMessage(ServerGamePacketListenerImpl.this.player.getUUID(), message);
+                            }
+                        } else {
+                            for (org.bukkit.entity.Player player : queueEvent.getRecipients()) {
+                                player.sendMessage(ServerGamePacketListenerImpl.this.player.getUUID(), message);
+                            }
+                        }
+                        ServerGamePacketListenerImpl.this.server.console.sendMessage(message);
+
+                        return null;
+                    }};
+                if (async) {
+                    server.processQueue.add(waitable);
+                } else {
+                    waitable.run();
+                }
+                try {
+                    waitable.get();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt(); // This is proper habit for java. If we aren't handling it, pass it on!
+                } catch (ExecutionException e) {
+                    throw new RuntimeException("Exception processing chat event", e.getCause());
+                }
+            } else {
+                if (event.isCancelled()) {
+                    return;
+                }
+
+                s = String.format(event.getFormat(), event.getPlayer().getDisplayName(), event.getMessage());
+                if (((LazyPlayerSet) event.getRecipients()).isLazy()) {
+                    if (originalFormat.equals(event.getFormat()) && originalMessage.equals(event.getMessage()) && event.getPlayer().getName().equalsIgnoreCase(event.getPlayer().getDisplayName())) {
+                        ServerGamePacketListenerImpl.this.server.getPlayerList().broadcastChatMessage(original, ServerGamePacketListenerImpl.this.player, ChatType.bind(ChatType.CHAT, (Entity) ServerGamePacketListenerImpl.this.player));
+                        return;
+                    }
+
+                    for (ServerPlayer recipient : server.getPlayerList().players) {
+                        recipient.getBukkitEntity().sendMessage(ServerGamePacketListenerImpl.this.player.getUUID(), s);
+                    }
+                } else {
+                    for (org.bukkit.entity.Player recipient : event.getRecipients()) {
+                        recipient.sendMessage(ServerGamePacketListenerImpl.this.player.getUUID(), s);
+                    }
+                }
+                server.console.sendMessage(s);
+            }
+        }
+    }
+
+    private void handleCommand(String s) {
+        this.LOGGER.info(this.player.getScoreboardName() + " issued server command: " + s);
+
+        CraftPlayer player = this.getCraftPlayer();
+
+        PlayerCommandPreprocessEvent event = new PlayerCommandPreprocessEvent(player, s, new LazyPlayerSet(server));
+        this.cserver.getPluginManager().callEvent(event);
+
+        if (event.isCancelled()) {
+            return;
+        }
+
+        try {
+            if (this.cserver.dispatchCommand(event.getPlayer(), event.getMessage().substring(1))) {
+                return;
+            }
+        } catch (org.bukkit.command.CommandException ex) {
+            player.sendMessage(org.bukkit.ChatColor.RED + "An internal error occurred while attempting to perform this command");
+            java.util.logging.Logger.getLogger(ServerGamePacketListenerImpl.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
+            return;
+        }
+    }
+    // CraftBukkit end
+
     private PlayerChatMessage getSignedMessage(ServerboundChatPacket packet, LastSeenMessages lastSeenMessages) throws SignedMessageChain.DecodeException {
         SignedMessageBody body = new SignedMessageBody(packet.message(), packet.timeStamp(), packet.salt(), lastSeenMessages);
         return this.signedMessageDecoder.unpack(packet.signature(), body);
@@ -1836,7 +_,11 @@
                     ItemStack tool = this.player.getItemInHand(hand);
                     if (tool.isItemEnabled(level.enabledFeatures())) {
                         ItemStack usedItemStack = tool.copy();
-                        if (this.player.interactOn(target, hand, location) instanceof InteractionResult.Success success) {
+                        var interactionResult = net.neoforged.neoforge.common.CommonHooks.onInteractEntityAt(player, target, location, hand);
+                        if (interactionResult == null) {
+                            interactionResult = this.player.interactOn(target, hand, location);
+                        }
+                        if (interactionResult instanceof InteractionResult.Success success) {
                             ItemStack awardedForStack = success.wasItemInteraction() ? usedItemStack : ItemStack.EMPTY;
                             CriteriaTriggers.PLAYER_INTERACTED_WITH_ENTITY.trigger(this.player, awardedForStack, target);
                             if (success.swingSource() == InteractionResult.SwingSource.SERVER) {
@@ -1889,7 +_,7 @@
                     this.restartClientLoadTimerAfterRespawn();
                     if (this.server.isHardcore()) {
                         this.player.setGameMode(GameType.SPECTATOR);
-                        this.player.level().getGameRules().set(GameRules.SPECTATORS_GENERATE_CHUNKS, false, this.server);
+                        this.player.level().getGameRules().set(GameRules.SPECTATORS_GENERATE_CHUNKS, false, this.player.level()); // CraftBukkit - per-world
                     }
                 }
                 break;
@@ -2057,14 +_,16 @@
     @Override
     public void handlePlayerAbilities(ServerboundPlayerAbilitiesPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
-        this.player.getAbilities().flying = packet.isFlying() && this.player.getAbilities().mayfly;
+        this.player.getAbilities().flying = packet.isFlying() && this.player.mayFly();
     }
 
     @Override
     public void handleClientInformation(ServerboundClientInformationPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.level());
         boolean wasHatShown = this.player.isModelPartShown(PlayerModelPart.HAT);
+        net.minecraft.server.level.ClientInformation oldInfo = this.player.clientInformation();
         this.player.updateOptions(packet.information());
+        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.ClientInformationUpdatedEvent(this.player, oldInfo, packet.information()));
         if (this.player.isModelPartShown(PlayerModelPart.HAT) != wasHatShown) {
             this.server.getPlayerList().broadcastAll(new ClientboundPlayerInfoUpdatePacket(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_HAT, this.player));
         }
@@ -2140,7 +_,7 @@
             this.connection
                 .setupInboundProtocol(
                     ConfigurationProtocols.SERVERBOUND,
-                    new ServerConfigurationPacketListenerImpl(this.server, this.connection, this.createCookie(this.player.clientInformation()))
+                    new ServerConfigurationPacketListenerImpl(this.server, this.connection, this.createCookie(this.player.clientInformation(), this.connectionType), this.player) // CraftBukkit
                 );
         }
     }
@@ -2175,6 +_,7 @@
 
     @Override
     public void handleCustomPayload(ServerboundCustomPayloadPacket packet) {
+        super.handleCustomPayload(packet); // Neo: Call super to invoke modded payload handling.
     }
 
     @Override
