--- a/net/minecraft/server/network/ServerTextFilter.java
+++ b/net/minecraft/server/network/ServerTextFilter.java
@@ -44,7 +_,7 @@
     private final ExecutorService workerPool;
 
     protected static ExecutorService createWorkerPool(int maxConcurrentRequests) {
-        return Executors.newFixedThreadPool(maxConcurrentRequests, THREAD_FACTORY);
+        return Executors.newVirtualThreadPerTaskExecutor();
     }
 
     protected ServerTextFilter(
