--- a/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
+++ b/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
@@ -42,7 +42,7 @@
     static CompletableFuture<ChunkAccess> generateStructureStarts(WorldGenContext worldgencontext, ChunkStep chunkstep, StaticCache2D<GenerationChunkHolder> staticcache2d, ChunkAccess chunkaccess) {
         ServerLevel serverlevel = worldgencontext.level();
 
-        if (serverlevel.getServer().getWorldData().worldGenOptions().generateStructures()) {
+        if (serverlevel.serverLevelData.worldGenOptions().generateStructures()) { // CraftBukkit
             worldgencontext.generator().createStructures(serverlevel.registryAccess(), serverlevel.getChunkSource().getGeneratorState(), serverlevel.structureManager(), chunkaccess, worldgencontext.structureManager(), serverlevel.dimension());
         }
 
