--- a/net/minecraft/client/gui/screens/options/controls/ControlsScreen.java
+++ b/net/minecraft/client/gui/screens/options/controls/ControlsScreen.java
@@ -1,14 +_,11 @@
 package net.minecraft.client.gui.screens.options.controls;
 
-import com.google.common.collect.Lists;
-import java.util.List;
 import net.minecraft.client.OptionInstance;
 import net.minecraft.client.Options;
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.client.gui.screens.options.MouseSettingsScreen;
 import net.minecraft.client.gui.screens.options.OptionsSubScreen;
-import net.minecraft.client.input.InputQuirks;
 import net.minecraft.network.chat.Component;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
@@ -18,21 +_,15 @@
     private static final Component TITLE = Component.translatable("controls.title");
 
     private static OptionInstance<?>[] options(Options options) {
-        List<OptionInstance<?>> result = Lists.newArrayList(
+        return new OptionInstance[]{
             options.toggleCrouch(),
             options.toggleSprint(),
             options.toggleAttack(),
             options.toggleUse(),
             options.autoJump(),
             options.sprintWindow(),
-            options.quitShortcuts()
-        );
-        if (InputQuirks.EMULATE_RIGHT_CLICK_WITH_CTRL_KEY) {
-            result.add(options.ctrlClickEmulatesRightClick());
-        }
-
-        result.add(options.operatorItemsTab());
-        return result.toArray(OptionInstance[]::new);
+            options.operatorItemsTab()
+        };
     }
 
     public ControlsScreen(Screen lastScreen, Options options) {
