--- a/net/minecraft/client/gui/screens/options/controls/KeyBindsList.java
+++ b/net/minecraft/client/gui/screens/options/controls/KeyBindsList.java
@@ -67,9 +_,7 @@
         private final FocusableTextWidget categoryName;
 
         public CategoryEntry(KeyMapping.Category category) {
-            this.categoryName = FocusableTextWidget.builder(
-                    category.label().copy().withStyle(ChatFormatting.UNDERLINE, ChatFormatting.BOLD), KeyBindsList.this.minecraft.font
-                )
+            this.categoryName = FocusableTextWidget.builder(category.label(), KeyBindsList.this.minecraft.font)
                 .alwaysShowBorder(false)
                 .backgroundFill(FocusableTextWidget.BackgroundFill.ON_FOCUS)
                 .build();
