--- a/com/mojang/realmsclient/dto/RealmsRegion.java
+++ b/com/mojang/realmsclient/dto/RealmsRegion.java
@@ -15,7 +_,6 @@
     AUSTRALIA_EAST("AustraliaEast", "realms.configuration.region.australia_east"),
     AUSTRALIA_SOUTHEAST("AustraliaSoutheast", "realms.configuration.region.australia_southeast"),
     BRAZIL_SOUTH("BrazilSouth", "realms.configuration.region.brazil_south"),
-    CANADA_CENTRAL("CanadaCentral", "realms.configuration.region.canada_central"),
     CENTRAL_INDIA("CentralIndia", "realms.configuration.region.central_india"),
     CENTRAL_US("CentralUs", "realms.configuration.region.central_us"),
     EAST_ASIA("EastAsia", "realms.configuration.region.east_asia"),
@@ -25,12 +_,10 @@
     JAPAN_EAST("JapanEast", "realms.configuration.region.japan_east"),
     JAPAN_WEST("JapanWest", "realms.configuration.region.japan_west"),
     KOREA_CENTRAL("KoreaCentral", "realms.configuration.region.korea_central"),
-    MEXICO_CENTRAL("MexicoCentral", "realms.configuration.region.mexico_central"),
     NORTH_CENTRAL_US("NorthCentralUs", "realms.configuration.region.north_central_us"),
     NORTH_EUROPE("NorthEurope", "realms.configuration.region.north_europe"),
     SOUTH_CENTRAL_US("SouthCentralUs", "realms.configuration.region.south_central_us"),
     SOUTHEAST_ASIA("SoutheastAsia", "realms.configuration.region.southeast_asia"),
-    SOUTH_AFRICA_NORTH("SouthAfricaNorth", "realms.configuration.region.south_africa_north"),
     SWEDEN_CENTRAL("SwedenCentral", "realms.configuration.region.sweden_central"),
     UAE_NORTH("UAENorth", "realms.configuration.region.uae_north"),
     UK_SOUTH("UKSouth", "realms.configuration.region.uk_south"),
@@ -38,7 +_,6 @@
     WEST_EUROPE("WestEurope", "realms.configuration.region.west_europe"),
     WEST_US("WestUs", "realms.configuration.region.west_us"),
     WEST_US_2("WestUs2", "realms.configuration.region.west_us_2"),
-    WEST_US_3("WestUs3", "realms.configuration.region.west_us_3"),
     INVALID_REGION("invalid", "");
 
     public final String nameId;
