--- a/com/mojang/realmsclient/dto/RealmsText.java
+++ b/com/mojang/realmsclient/dto/RealmsText.java
@@ -33,7 +_,7 @@
         if (!I18n.exists(this.translationKey)) {
             return null;
         } else {
-            return this.args == null ? Component.translatable(this.translationKey) : Component.translatable(this.translationKey, this.args);
+            return this.args == null ? Component.translatable(this.translationKey) : Component.translatable(this.translationKey, (Object) this.args);
         }
     }
 
