--- a/net/minecraft/client/renderer/fog/environment/FogEnvironment.java
+++ b/net/minecraft/client/renderer/fog/environment/FogEnvironment.java
@@ -4,27 +_,23 @@
 import net.minecraft.client.DeltaTracker;
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.client.renderer.fog.FogData;
-import net.minecraft.util.ARGB;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.level.material.FogType;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
-import org.joml.Vector3fc;
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
 public abstract class FogEnvironment {
-    private static final Vector3fc WHITE = ARGB.vector3fFromRGB24(-1);
-
     public abstract void setupFog(FogData fog, Camera camera, ClientLevel level, float renderDistance, DeltaTracker deltaTracker);
 
     public boolean providesColor() {
         return true;
     }
 
-    public Vector3fc getBaseColor(ClientLevel level, Camera camera, int renderDistance, float partialTicks) {
-        return WHITE;
+    public int getBaseColor(ClientLevel level, Camera camera, int renderDistance, float partialTicks) {
+        return -1;
     }
 
     public boolean modifiesDarkness() {
