--- a/net/minecraft/client/renderer/fog/environment/PowderedSnowFogEnvironment.java
+++ b/net/minecraft/client/renderer/fog/environment/PowderedSnowFogEnvironment.java
@@ -4,21 +_,19 @@
 import net.minecraft.client.DeltaTracker;
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.client.renderer.fog.FogData;
-import net.minecraft.util.ARGB;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.material.FogType;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
-import org.joml.Vector3fc;
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
 public class PowderedSnowFogEnvironment extends FogEnvironment {
-    private static final Vector3fc COLOR = ARGB.vector3fFromRGB24(-6308916);
+    private static final int COLOR = -6308916;
 
     @Override
-    public Vector3fc getBaseColor(ClientLevel level, Camera camera, int renderDistance, float partialTicks) {
-        return COLOR;
+    public int getBaseColor(ClientLevel level, Camera camera, int renderDistance, float partialTicks) {
+        return -6308916;
     }
 
     @Override
