--- a/net/minecraft/client/renderer/fog/environment/WaterFogEnvironment.java
+++ b/net/minecraft/client/renderer/fog/environment/WaterFogEnvironment.java
@@ -10,7 +_,6 @@
 import net.minecraft.world.level.material.FogType;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
-import org.joml.Vector3fc;
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
@@ -34,7 +_,7 @@
     }
 
     @Override
-    public Vector3fc getBaseColor(ClientLevel level, Camera camera, int renderDistance, float partialTicks) {
+    public int getBaseColor(ClientLevel level, Camera camera, int renderDistance, float partialTicks) {
         return camera.attributeProbe().getValue(EnvironmentAttributes.WATER_FOG_COLOR, partialTicks);
     }
 }
