--- a/net/minecraft/world/damagesource/DamageType.java
+++ b/net/minecraft/world/damagesource/DamageType.java
@@ -4,10 +_,10 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.RegistryFixedCodec;
 
 public record DamageType(String msgId, DamageScaling scaling, float exhaustion, DamageEffects effects, DeathMessageType deathMessageType) {
     public static final Codec<DamageType> DIRECT_CODEC = RecordCodecBuilder.create(
@@ -20,7 +_,7 @@
             )
             .apply(i, DamageType::new)
     );
-    public static final Codec<Holder<DamageType>> CODEC = RegistryCodecs.holder(Registries.DAMAGE_TYPE);
+    public static final Codec<Holder<DamageType>> CODEC = RegistryFixedCodec.create(Registries.DAMAGE_TYPE);
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<DamageType>> STREAM_CODEC = ByteBufCodecs.holderRegistry(Registries.DAMAGE_TYPE);
 
     public DamageType(String msgdId, DamageScaling scaling, float exhaustion) {
