--- a/net/minecraft/world/damagesource/DamageSource.java
+++ b/net/minecraft/world/damagesource/DamageSource.java
@@ -18,6 +_,124 @@
     private final @Nullable Entity directEntity;
     private final @Nullable Vec3 damageSourcePosition;
 
+    // CraftBukkit start
+    private org.bukkit.event.entity.EntityDamageEvent.@Nullable DamageCause knownCause; // When the damage event cause is known by the context of the call rather than the damage source data
+    @Nullable
+    private Entity eventEntityDamager = null; // Relevant entity set when the game doesn't normally set a causingEntity/directEntity
+    private org.bukkit.block.@Nullable Block eventBlockDamager; // Relevant block set. damageSourcePosition is only used for bad respawn point explosion or custom damage
+    private org.bukkit.block.@Nullable BlockState fromBlockSnapshot; // Captured block snapshot when the eventBlockDamager is not relevant (e.g. for bad respawn point explosions the block is already removed)
+    private boolean critical; // Supports arrows and sweeping damage
+    private final net.kyori.adventure.pointer.Pointers damageContext;
+    private boolean scissors = false; // Purpur - Dont run with scissors!
+    private boolean stonecutter = false; // Purpur - Stonecutter damage
+
+    public DamageSource knownCause(final org.bukkit.event.entity.EntityDamageEvent.DamageCause cause) {
+        final DamageSource damageSource = this.copy();
+        damageSource.knownCause = cause;
+        return damageSource;
+    }
+
+    public org.bukkit.event.entity.EntityDamageEvent.@Nullable DamageCause knownCause() {
+        return this.knownCause;
+    }
+
+    // Purpur start - Dont run with scissors!
+    public DamageSource scissors() {
+        this.knownCause(org.bukkit.event.entity.EntityDamageEvent.DamageCause.SUICIDE);
+        this.scissors = true;
+        return this;
+    }
+
+    public boolean isScissors() {
+        return this.scissors;
+    }
+    // Purpur end - Dont run with scissors!
+
+    // Purpur start -  - Stonecutter damage
+    public DamageSource stonecutter() {
+        this.knownCause(org.bukkit.event.entity.EntityDamageEvent.DamageCause.CONTACT);
+        this.stonecutter = true;
+        return this;
+    }
+
+    public boolean isStonecutter() {
+        return this.stonecutter;
+    }
+    // Purpur end - Stonecutter damage
+
+    @Nullable
+    public Entity eventEntityDamager() {
+        return this.eventEntityDamager;
+    }
+
+    public DamageSource eventEntityDamager(final Entity entity) {
+        if (this.directEntity == entity) { // no changes needed but useful for data driven damage
+            return this;
+        }
+
+        if (this.directEntity != null) {
+            throw new IllegalStateException("Cannot set an event damager when another direct entity is already set (report a bug to Paper)");
+        }
+        final DamageSource damageSource = this.copy();
+        damageSource.eventEntityDamager = entity;
+        return damageSource;
+    }
+
+    public org.bukkit.block.@Nullable Block eventBlockDamager() {
+        return this.eventBlockDamager;
+    }
+
+    public DamageSource eventBlockDamager(final net.minecraft.world.level.LevelAccessor level, final net.minecraft.core.@Nullable BlockPos pos) {
+        if (pos == null) {
+            return this;
+        }
+
+        final DamageSource damageSource = this.copy();
+        damageSource.eventBlockDamager = org.bukkit.craftbukkit.block.CraftBlock.at(level, pos);
+        return damageSource;
+    }
+
+    public org.bukkit.block.@Nullable BlockState causingBlockSnapshot() {
+        return this.fromBlockSnapshot;
+    }
+
+    public DamageSource causingBlockSnapshot(final org.bukkit.block.BlockState blockState) {
+        if (this.eventBlockDamager != null) {
+            throw new IllegalStateException("Cannot set a block snapshot when an event block damager is already set (report a bug to Paper)");
+        }
+        final DamageSource damageSource = this.copy();
+        damageSource.fromBlockSnapshot = blockState;
+        return damageSource;
+    }
+
+    public boolean isCritical() {
+        return this.critical;
+    }
+
+    public DamageSource critical() {
+        final DamageSource damageSource = this.copy();
+        damageSource.critical = true;
+        return damageSource;
+    }
+
+    public net.kyori.adventure.pointer.Pointers getDamageContext() {
+        return this.damageContext;
+    }
+
+    // Cloning the instance lets us return unique instances of DamageSource without affecting constants defined in DamageSources
+    private DamageSource copy() {
+        final DamageSource damageSource = new DamageSource(this.type, this.directEntity, this.causingEntity, this.damageSourcePosition, this.damageContext);
+        damageSource.knownCause = this.knownCause;
+        damageSource.eventEntityDamager = this.eventEntityDamager;
+        damageSource.eventBlockDamager = this.eventBlockDamager;
+        damageSource.fromBlockSnapshot = this.fromBlockSnapshot;
+        damageSource.critical = this.critical;
+        damageSource.scissors = this.isScissors(); // Purpur - Dont run with scissors!
+        damageSource.stonecutter = this.isStonecutter(); // Purpur - Stonecutter damage
+        return damageSource;
+    }
+    // CraftBukkit end
+
     @Override
     public String toString() {
         return "DamageSource (" + this.type().msgId() + ")";
@@ -31,15 +_,22 @@
         return this.causingEntity == this.directEntity;
     }
 
-    public DamageSource(Holder<DamageType> type, @Nullable Entity directEntity, @Nullable Entity causingEntity, @Nullable Vec3 damageSourcePosition) {
+    public DamageSource(Holder<DamageType> type, @Nullable Entity directEntity, @Nullable Entity causingEntity, @Nullable Vec3 damageSourcePosition, final net.kyori.adventure.pointer.Pointers damageContext) { // Paper - Add damage context
         this.type = type;
         this.causingEntity = causingEntity;
         this.directEntity = directEntity;
         this.damageSourcePosition = damageSourcePosition;
-    }
+        // Paper start - Add damage context
+        this.damageContext = damageContext;
+    }
+
+    public DamageSource(Holder<DamageType> type, @Nullable Entity directEntity, @Nullable Entity causingEntity, @Nullable Vec3 damageSourcePosition) {
+        this(type, directEntity, causingEntity, damageSourcePosition, net.kyori.adventure.pointer.Pointers.empty());
+    }
+    // Paper end - Add damage context
 
     public DamageSource(Holder<DamageType> type, @Nullable Entity directEntity, @Nullable Entity causingEntity) {
-        this(type, directEntity, causingEntity, null);
+        this(type, directEntity, causingEntity, null, net.kyori.adventure.pointer.Pointers.empty());
     }
 
     public DamageSource(Holder<DamageType> type, Vec3 damageSourcePosition) {
@@ -77,16 +_,29 @@
         } else {
             Component name = this.causingEntity == null ? this.directEntity.getDisplayName() : this.causingEntity.getDisplayName();
             ItemStack held = this.causingEntity instanceof LivingEntity livingEntity ? livingEntity.getMainHandItem() : ItemStack.EMPTY;
-            return !held.isEmpty() && held.has(DataComponents.CUSTOM_NAME)
+            return !held.isEmpty() && (org.purpurmc.purpur.PurpurConfig.playerDeathsAlwaysShowItem || held.has(DataComponents.CUSTOM_NAME)) // Purpur - always show item in player death messages
                 ? Component.translatable(deathMsg + ".item", victim.getDisplayName(), name, held.getDisplayName())
                 : Component.translatable(deathMsg, victim.getDisplayName(), name);
         }
     }
 
+    // Purpur start - Component related conveniences
+    public Component getLocalizedDeathMessage(String str, LivingEntity entity) {
+        net.kyori.adventure.text.Component name = io.papermc.paper.adventure.PaperAdventure.asAdventure(entity.getDisplayName());
+        net.kyori.adventure.text.minimessage.tag.resolver.TagResolver template = net.kyori.adventure.text.minimessage.tag.resolver.Placeholder.component("player", name);
+        net.kyori.adventure.text.Component component = net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(str, template);
+        return io.papermc.paper.adventure.PaperAdventure.asVanilla(component);
+    }
+    // Purpur end - Component related conveniences
+
     public String getMsgId() {
         return this.type().msgId();
     }
 
+    /**
+     * @deprecated Use {@link DamageScaling#getScalingFunction()}
+     */
+    @Deprecated(since = "1.20.1")
     public boolean scalesWithDifficulty() {
         return switch (this.type().scaling()) {
             case NEVER -> false;
