--- a/net/minecraft/data/registries/RegistryPatchGenerator.java
+++ b/net/minecraft/data/registries/RegistryPatchGenerator.java
@@ -1,53 +_,40 @@
 package net.minecraft.data.registries;
 
+import com.mojang.datafixers.DataFixUtils;
+import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import net.minecraft.core.Cloner;
 import net.minecraft.core.HolderLookup;
-import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.RegistrySetBuilder;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.resources.RegistryDataLoader;
-import net.minecraft.resources.ResourceKey;
-import net.minecraft.world.level.storage.loot.LootDataType;
+import net.minecraft.world.level.biome.Biome;
+import net.minecraft.world.level.levelgen.placement.PlacedFeature;
 
 public class RegistryPatchGenerator {
-    private static boolean hasAnyPatchedElement(RegistrySetBuilder.PatchedRegistries newRegistries, ResourceKey<? extends Registry<?>> registry) {
-        return newRegistries.patches().lookup(registry).flatMap(lookup -> lookup.listElements().findAny()).isPresent();
-    }
-
-    public static CompletableFuture<RegistrySetBuilder.PatchedRegistries> createWorldLookup(
+    public static CompletableFuture<RegistrySetBuilder.PatchedRegistries> createLookup(
         CompletableFuture<HolderLookup.Provider> vanilla, RegistrySetBuilder packBuilder
     ) {
-        return vanilla.thenApply(parent -> {
-            RegistryAccess.Frozen staticRegistries = RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY);
-            Cloner.Factory cloner = new Cloner.Factory();
-            RegistryDataLoader.WORLD_REGISTRIES.forEach(registryData -> registryData.runWithArguments(cloner::addCodec));
-            RegistrySetBuilder.PatchedRegistries newRegistries = packBuilder.buildPatch(staticRegistries, parent, cloner);
-            boolean hasAnyPatchedBiomes = hasAnyPatchedElement(newRegistries, Registries.BIOME);
-            boolean hasAnyPatchedFeatures = hasAnyPatchedElement(newRegistries, Registries.PLACED_FEATURE);
-            if (hasAnyPatchedBiomes || hasAnyPatchedFeatures) {
-                VanillaRegistries.validateThatAllBiomeFeaturesHaveBiomeFilter(newRegistries.full());
-            }
-
-            return newRegistries;
-        });
-    }
-
-    public static CompletableFuture<RegistrySetBuilder.PatchedRegistries> createReloadableLookup(
-        CompletableFuture<HolderLookup.Provider> contextFuture, CompletableFuture<HolderLookup.Provider> vanillaFuture, RegistrySetBuilder packBuilder
-    ) {
-        return contextFuture.thenCombine(vanillaFuture, (context, vanilla) -> {
-            Cloner.Factory cloner = new Cloner.Factory();
-            RegistryDataLoader.RELOADABLE_REGISTRIES.forEach(registryData -> registryData.runWithArguments(cloner::addCodec));
-            RegistrySetBuilder.PatchedRegistries newRegistries = packBuilder.buildPatch(context, vanilla, cloner);
-            boolean hasAnyPatchedLootData = LootDataType.values().anyMatch(type -> hasAnyPatchedElement(newRegistries, type.registryKey()));
-            if (hasAnyPatchedLootData) {
-                VanillaRegistries.validateLootData(newRegistries.full());
-            }
-
-            return newRegistries;
-        });
+        return vanilla.thenApply(
+            parent -> {
+                RegistryAccess.Frozen staticRegistries = RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY);
+                Cloner.Factory cloner = new Cloner.Factory();
+                net.neoforged.neoforge.registries.DataPackRegistriesHooks.getDataPackRegistriesWithDimensions().forEach(registryData -> registryData.runWithArguments(cloner::addCodec));
+                RegistrySetBuilder.PatchedRegistries newRegistries = packBuilder.buildPatch(staticRegistries, parent, cloner);
+                HolderLookup.Provider fullPatchedRegistry = newRegistries.full();
+                Optional<? extends HolderLookup.RegistryLookup<Biome>> biomes = fullPatchedRegistry.lookup(Registries.BIOME);
+                Optional<? extends HolderLookup.RegistryLookup<PlacedFeature>> features = fullPatchedRegistry.lookup(Registries.PLACED_FEATURE);
+                if (biomes.isPresent() || features.isPresent()) {
+                    VanillaRegistries.validateThatAllBiomeFeaturesHaveBiomeFilter(
+                        DataFixUtils.orElseGet(features, () -> parent.lookupOrThrow(Registries.PLACED_FEATURE)),
+                        DataFixUtils.orElseGet(biomes, () -> parent.lookupOrThrow(Registries.BIOME))
+                    );
+                }
+
+                return newRegistries;
+            }
+        );
     }
 }
