--- a/net/minecraft/data/registries/VanillaRegistries.java
+++ b/net/minecraft/data/registries/VanillaRegistries.java
@@ -1,17 +_,14 @@
 package net.minecraft.data.registries;
 
-import com.mojang.logging.LogUtils;
 import java.util.List;
 import net.minecraft.core.Holder;
+import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderLookup;
 import net.minecraft.core.HolderSet;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.RegistrySetBuilder;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.data.advancements.packs.VanillaAdvancementProvider;
-import net.minecraft.data.loot.packs.VanillaLootTableProvider;
-import net.minecraft.data.recipes.packs.VanillaRecipeProvider;
 import net.minecraft.data.worldgen.Carvers;
 import net.minecraft.data.worldgen.DimensionTypes;
 import net.minecraft.data.worldgen.NoiseData;
@@ -21,8 +_,6 @@
 import net.minecraft.data.worldgen.Structures;
 import net.minecraft.data.worldgen.biome.BiomeData;
 import net.minecraft.data.worldgen.features.FeatureUtils;
-import net.minecraft.data.worldgen.material.VanillaMaterialConditions;
-import net.minecraft.data.worldgen.material.VanillaMaterialRules;
 import net.minecraft.data.worldgen.placement.PlacementUtils;
 import net.minecraft.gametest.framework.GameTestEnvironments;
 import net.minecraft.gametest.framework.GameTestInstances;
@@ -30,7 +_,6 @@
 import net.minecraft.resources.Identifier;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.dialog.Dialogs;
-import net.minecraft.util.ProblemReporter;
 import net.minecraft.util.Util;
 import net.minecraft.world.clock.WorldClocks;
 import net.minecraft.world.damagesource.DamageTypes;
@@ -59,7 +_,6 @@
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.MultiNoiseBiomeSourceParameterLists;
 import net.minecraft.world.level.block.entity.BannerPatterns;
-import net.minecraft.world.level.block.entity.DecoratedPotPatterns;
 import net.minecraft.world.level.block.entity.trialspawner.TrialSpawnerConfigs;
 import net.minecraft.world.level.levelgen.NoiseGeneratorSettings;
 import net.minecraft.world.level.levelgen.NoiseRouterData;
@@ -67,19 +_,13 @@
 import net.minecraft.world.level.levelgen.placement.BiomeFilter;
 import net.minecraft.world.level.levelgen.placement.PlacedFeature;
 import net.minecraft.world.level.levelgen.presets.WorldPresets;
-import net.minecraft.world.level.storage.loot.LootDataType;
-import net.minecraft.world.level.storage.loot.ValidationContextSource;
-import net.minecraft.world.level.storage.loot.predicates.LootPredicates;
-import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 import net.minecraft.world.timeline.Timelines;
-import org.slf4j.Logger;
 
 public class VanillaRegistries {
-    private static final Logger LOGGER = LogUtils.getLogger();
-    private static final RegistrySetBuilder WORLD_BUILDER = new RegistrySetBuilder()
+    private static final RegistrySetBuilder BUILDER = new RegistrySetBuilder()
         .add(Registries.DIMENSION_TYPE, DimensionTypes::bootstrap)
-        .add(Registries.CARVER, Carvers::bootstrap)
-        .add(Registries.FEATURE, FeatureUtils::bootstrap)
+        .add(Registries.CONFIGURED_CARVER, Carvers::bootstrap)
+        .add(Registries.CONFIGURED_FEATURE, FeatureUtils::bootstrap)
         .add(Registries.PLACED_FEATURE, PlacementUtils::bootstrap)
         .add(Registries.STRUCTURE, Structures::bootstrap)
         .add(Registries.STRUCTURE_SET, StructureSets::bootstrap)
@@ -89,8 +_,6 @@
         .add(Registries.MULTI_NOISE_BIOME_SOURCE_PARAMETER_LIST, MultiNoiseBiomeSourceParameterLists::bootstrap)
         .add(Registries.NOISE, NoiseData::bootstrap)
         .add(Registries.DENSITY_FUNCTION, NoiseRouterData::bootstrap)
-        .add(Registries.MATERIAL_RULE, VanillaMaterialRules::bootstrap)
-        .add(Registries.MATERIAL_CONDITION, VanillaMaterialConditions::bootstrap)
         .add(Registries.NOISE_SETTINGS, NoiseGeneratorSettings::bootstrap)
         .add(Registries.WORLD_PRESET, WorldPresets::bootstrap)
         .add(Registries.FLAT_LEVEL_GENERATOR_PRESET, FlatLevelGeneratorPresets::bootstrap)
@@ -124,18 +_,16 @@
         .add(Registries.WORLD_CLOCK, WorldClocks::bootstrap)
         .add(Registries.TIMELINE, Timelines::bootstrap)
         .add(Registries.VILLAGER_TRADE, VillagerTrades::bootstrap)
-        .add(Registries.TRADE_SET, TradeSets::bootstrap)
-        .add(Registries.DECORATED_POT_PATTERN, DecoratedPotPatterns::bootstrap);
-    private static final RegistrySetBuilder RELOADABLE_BUILDER = new RegistrySetBuilder()
-        .add(Registries.LOOT_TABLE, VanillaLootTableProvider.create())
-        .add(Registries.PREDICATE, LootPredicates::bootstrap)
-        .add(Registries.ADVANCEMENT, VanillaAdvancementProvider.create())
-        .add(Registries.NUMBER_PROVIDER, NumberProviders::bootstrap)
-        .add(VanillaRecipeProvider.create());
-
-    public static void validateThatAllBiomeFeaturesHaveBiomeFilter(HolderLookup.Provider registries) {
-        HolderLookup.RegistryLookup<Biome> biomes = registries.lookupOrThrow(Registries.BIOME);
-        HolderLookup.RegistryLookup<PlacedFeature> placedFeatures = registries.lookupOrThrow(Registries.PLACED_FEATURE);
+        .add(Registries.TRADE_SET, TradeSets::bootstrap);
+
+    // Neo: Keep a set of all actual vanilla registries to allow filtering out registries that are added by mods for Vanilla clients.
+    public static final List<? extends net.minecraft.resources.ResourceKey<? extends net.minecraft.core.Registry<?>>> DATAPACK_REGISTRY_KEYS = BUILDER.getEntryKeys();
+
+    private static void validateThatAllBiomeFeaturesHaveBiomeFilter(HolderLookup.Provider provider) {
+        validateThatAllBiomeFeaturesHaveBiomeFilter(provider.lookupOrThrow(Registries.PLACED_FEATURE), provider.lookupOrThrow(Registries.BIOME));
+    }
+
+    public static void validateThatAllBiomeFeaturesHaveBiomeFilter(HolderGetter<PlacedFeature> placedFeatures, HolderLookup<Biome> biomes) {
         biomes.listElements().forEach(biome -> {
             Identifier biomeKey = biome.key().identifier();
             List<HolderSet<PlacedFeature>> biomeFeatures = biome.value().getGenerationSettings().features();
@@ -156,26 +_,10 @@
         return value.placement().contains(BiomeFilter.biome());
     }
 
-    public static void validateLootData(HolderLookup.Provider newRegistries) {
-        ProblemReporter.Collector problems = new ProblemReporter.Collector();
-        ValidationContextSource validationContext = new ValidationContextSource(problems, newRegistries);
-        LootDataType.values().forEach(lootDataType -> lootDataType.runValidationIfPresent(validationContext, newRegistries));
-        if (!problems.isEmpty()) {
-            problems.forEach((id, problem) -> LOGGER.warn("Found validation problem in {}: {}", id, problem.description()));
-            throw new IllegalStateException("Failed to validate loot data, see logs");
-        }
-    }
-
-    public static HolderLookup.Provider createWorldLookup() {
+    public static HolderLookup.Provider createLookup() {
         RegistryAccess.Frozen staticRegistries = RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY);
-        HolderLookup.Provider newRegistries = WORLD_BUILDER.build(staticRegistries);
+        HolderLookup.Provider newRegistries = BUILDER.build(staticRegistries);
         validateThatAllBiomeFeaturesHaveBiomeFilter(newRegistries);
-        return newRegistries;
-    }
-
-    public static HolderLookup.Provider createReloadableLookup(HolderLookup.Provider context) {
-        HolderLookup.Provider newRegistries = RELOADABLE_BUILDER.build(context);
-        validateLootData(newRegistries);
         return newRegistries;
     }
 }
