--- a/net/minecraft/data/registries/VanillaRegistries.java
+++ b/net/minecraft/data/registries/VanillaRegistries.java
@@ -18,8 +_,6 @@
 import net.minecraft.data.worldgen.Structures;
 import net.minecraft.data.worldgen.biome.BiomeData;
 import net.minecraft.data.worldgen.features.FeatureUtils;
-import net.minecraft.data.worldgen.material.VanillaMaterialConditions;
-import net.minecraft.data.worldgen.material.VanillaMaterialRules;
 import net.minecraft.data.worldgen.placement.PlacementUtils;
 import net.minecraft.gametest.framework.GameTestEnvironments;
 import net.minecraft.gametest.framework.GameTestInstances;
@@ -55,7 +_,6 @@
 import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.MultiNoiseBiomeSourceParameterLists;
 import net.minecraft.world.level.block.entity.BannerPatterns;
-import net.minecraft.world.level.block.entity.DecoratedPotPatterns;
 import net.minecraft.world.level.block.entity.trialspawner.TrialSpawnerConfigs;
 import net.minecraft.world.level.levelgen.NoiseGeneratorSettings;
 import net.minecraft.world.level.levelgen.NoiseRouterData;
@@ -68,8 +_,8 @@
 public class VanillaRegistries {
     private static final RegistrySetBuilder BUILDER = new RegistrySetBuilder()
         .add(Registries.DIMENSION_TYPE, DimensionTypes::bootstrap)
-        .add(Registries.CARVER, Carvers::bootstrap)
-        .add(Registries.FEATURE, FeatureUtils::bootstrap)
+        .add(Registries.CONFIGURED_CARVER, Carvers::bootstrap)
+        .add(Registries.CONFIGURED_FEATURE, FeatureUtils::bootstrap)
         .add(Registries.PLACED_FEATURE, PlacementUtils::bootstrap)
         .add(Registries.STRUCTURE, Structures::bootstrap)
         .add(Registries.STRUCTURE_SET, StructureSets::bootstrap)
@@ -79,8 +_,6 @@
         .add(Registries.MULTI_NOISE_BIOME_SOURCE_PARAMETER_LIST, MultiNoiseBiomeSourceParameterLists::bootstrap)
         .add(Registries.NOISE, NoiseData::bootstrap)
         .add(Registries.DENSITY_FUNCTION, NoiseRouterData::bootstrap)
-        .add(Registries.MATERIAL_RULE, VanillaMaterialRules::bootstrap)
-        .add(Registries.MATERIAL_CONDITION, VanillaMaterialConditions::bootstrap)
         .add(Registries.NOISE_SETTINGS, NoiseGeneratorSettings::bootstrap)
         .add(Registries.WORLD_PRESET, WorldPresets::bootstrap)
         .add(Registries.FLAT_LEVEL_GENERATOR_PRESET, FlatLevelGeneratorPresets::bootstrap)
@@ -114,8 +_,10 @@
         .add(Registries.WORLD_CLOCK, WorldClocks::bootstrap)
         .add(Registries.TIMELINE, Timelines::bootstrap)
         .add(Registries.VILLAGER_TRADE, VillagerTrades::bootstrap)
-        .add(Registries.TRADE_SET, TradeSets::bootstrap)
-        .add(Registries.DECORATED_POT_PATTERN, DecoratedPotPatterns::bootstrap);
+        .add(Registries.TRADE_SET, TradeSets::bootstrap);
+
+    // Neo: Keep a set of all actual vanilla registries to allow filtering out registries that are added by mods for Vanilla clients.
+    public static final List<? extends net.minecraft.resources.ResourceKey<? extends net.minecraft.core.Registry<?>>> DATAPACK_REGISTRY_KEYS = BUILDER.getEntryKeys();
 
     private static void validateThatAllBiomeFeaturesHaveBiomeFilter(HolderLookup.Provider provider) {
         validateThatAllBiomeFeaturesHaveBiomeFilter(provider.lookupOrThrow(Registries.PLACED_FEATURE), provider.lookupOrThrow(Registries.BIOME));
