--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
@@ -41,7 +_,7 @@
     public void doServerTick() {
         if (this.targetLocation == null) {
             this.targetLocation = Vec3.atBottomCenterOf(
-                this.dragon.level().getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, EndPodiumFeature.getLocation(this.dragon.getFightOrigin()))
+                this.dragon.level().getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()) // Paper - Allow changing the EnderDragon podium
             );
         }
 
