--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonStrafePlayerPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonStrafePlayerPhase.java
@@ -78,8 +_,11 @@
                         }
 
                         DragonFireball dragonfireball = new DragonFireball(this.dragon.level(), this.dragon, vec31.normalize());
+                        dragonfireball.preserveMotion = true; // Paper - Fix Entity Teleportation and cancel velocity if teleported
                         dragonfireball.moveTo(d6, d7, d8, 0.0F, 0.0F);
-                        this.dragon.level().addFreshEntity(dragonfireball);
+                        if (new com.destroystokyo.paper.event.entity.EnderDragonShootFireballEvent((org.bukkit.entity.EnderDragon) dragon.getBukkitEntity(), (org.bukkit.entity.DragonFireball) dragonfireball.getBukkitEntity()).callEvent()) // Paper - EnderDragon Events
+                            this.dragon.level().addFreshEntity(dragonfireball);
+                        else dragonfireball.discard(null); // Paper - EnderDragon Events
                         this.fireballCharge = 0;
                         if (this.currentPath != null) {
                             while (!this.currentPath.isDone()) {
