--- a/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
+++ b/net/minecraft/world/entity/boss/enderdragon/phases/DragonLandingPhase.java
@@ -5,8 +_,8 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.entity.boss.enderdragon.EnderDragon;
-import net.minecraft.world.level.dimension.end.EnderDragonFight;
 import net.minecraft.world.level.levelgen.Heightmap;
+import net.minecraft.world.level.levelgen.feature.EndPodiumFeature;
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
@@ -50,7 +_,7 @@
     public void doServerTick(ServerLevel level) {
         if (this.targetLocation == null) {
             this.targetLocation = Vec3.atBottomCenterOf(
-                level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, EnderDragonFight.getPodiumLocation(this.dragon.getFightOrigin()))
+                level.getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, this.dragon.getPodium()) // Paper - Allow changing the EnderDragon podium
             );
         }
 
