--- a/net/minecraft/data/recipes/RecipeUnlockAdvancementBuilder.java
+++ b/net/minecraft/data/recipes/RecipeUnlockAdvancementBuilder.java
@@ -8,7 +_,6 @@
 import net.minecraft.advancements.AdvancementRewards;
 import net.minecraft.advancements.triggers.Criterion;
 import net.minecraft.advancements.triggers.RecipeUnlockedTrigger;
-import net.minecraft.core.registries.Registries;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.world.item.crafting.Recipe;
 
@@ -20,15 +_,19 @@
     }
 
     public AdvancementHolder build(RecipeOutput output, ResourceKey<Recipe<?>> id, RecipeCategory category) {
+        return this.build(output, id, category.getFolderName());
+    }
+
+    public AdvancementHolder build(RecipeOutput output, ResourceKey<Recipe<?>> id, @org.jspecify.annotations.Nullable String category) {
         if (this.criteria.isEmpty()) {
             throw new IllegalStateException("No way of obtaining recipe " + id.identifier());
         }
 
         Advancement.Builder advancement = output.advancement()
-            .addCriterion("has_the_recipe", RecipeUnlockedTrigger.unlocked(output.lookup(Registries.RECIPE).getOrThrow(id)))
+            .addCriterion("has_the_recipe", RecipeUnlockedTrigger.unlocked(id))
             .rewards(AdvancementRewards.Builder.recipe(id))
             .requirements(AdvancementRequirements.Strategy.OR);
         this.criteria.forEach(advancement::addCriterion);
-        return advancement.build(id.identifier().withPrefix("recipes/" + category.getFolderName() + "/"));
+        return advancement.build(id.identifier().withPrefix("recipes/" + (category != null ? category + "/" : "")));
     }
 }
