--- a/net/minecraft/data/recipes/RecipeOutput.java
+++ b/net/minecraft/data/recipes/RecipeOutput.java
@@ -2,15 +_,18 @@
 
 import net.minecraft.advancements.Advancement;
 import net.minecraft.advancements.AdvancementHolder;
-import net.minecraft.data.worldgen.BootstrapContextAccess;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.world.item.crafting.Recipe;
 import org.jspecify.annotations.Nullable;
 
 import net.neoforged.neoforge.common.extensions.IRecipeOutputExtension;
 
-public interface RecipeOutput extends BootstrapContextAccess, IRecipeOutputExtension {
-    void accept(ResourceKey<Recipe<?>> id, Recipe<?> recipe, @Nullable AdvancementHolder advancement);
+public interface RecipeOutput extends IRecipeOutputExtension {
+    default void accept(ResourceKey<Recipe<?>> id, Recipe<?> recipe, @Nullable AdvancementHolder advancement) {
+        accept(id, recipe, advancement, new net.neoforged.neoforge.common.conditions.ICondition[0]);
+    }
 
     Advancement.Builder advancement();
+
+    void includeRootAdvancement();
 }
