--- a/net/minecraft/server/dialog/Dialog.java
+++ b/net/minecraft/server/dialog/Dialog.java
@@ -6,20 +_,21 @@
 import java.util.Optional;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.RegistryFileCodec;
 import net.minecraft.server.dialog.action.Action;
 import net.minecraft.util.ExtraCodecs;
 
 public interface Dialog {
     Codec<Integer> WIDTH_CODEC = ExtraCodecs.intRange(1, 1024);
     Codec<Dialog> DIRECT_CODEC = BuiltInRegistries.DIALOG_TYPE.byNameCodec().dispatch(Dialog::codec, c -> c);
-    Codec<Holder<Dialog>> CODEC = RegistryCodecs.holder(Registries.DIALOG, DIRECT_CODEC);
-    Codec<HolderSet<Dialog>> LIST_CODEC = RegistryCodecs.holderSet(Registries.DIALOG, DIRECT_CODEC);
+    Codec<Holder<Dialog>> CODEC = RegistryFileCodec.create(Registries.DIALOG, DIRECT_CODEC);
+    Codec<HolderSet<Dialog>> LIST_CODEC = RegistryCodecs.homogeneousList(Registries.DIALOG, DIRECT_CODEC);
     StreamCodec<RegistryFriendlyByteBuf, Holder<Dialog>> STREAM_CODEC = ByteBufCodecs.holder(
         Registries.DIALOG, ByteBufCodecs.fromCodecWithRegistriesTrusted(DIRECT_CODEC)
     );
