--- a/net/minecraft/client/server/IntegratedPlayerList.java
+++ b/net/minecraft/client/server/IntegratedPlayerList.java
@@ -18,10 +_,8 @@
     }
 
     @Override
-    public Component canPlayerLogin(SocketAddress address, NameAndId nameAndId) {
-        return this.getServer().isSingleplayerOwner(nameAndId) && this.getPlayerByName(nameAndId.name()) != null
-            ? Component.translatable("multiplayer.disconnect.name_taken")
-            : super.canPlayerLogin(address, nameAndId);
+    public LoginResult canPlayerLogin(SocketAddress address, NameAndId nameAndId) {
+        return super.canPlayerLogin(address, nameAndId);
     }
 
     public IntegratedServer getServer() {
