--- a/net/minecraft/client/server/IntegratedServer.java
+++ b/net/minecraft/client/server/IntegratedServer.java
@@ -79,7 +_,7 @@
         Services services,
         LevelLoadListener levelLoadListener
     ) {
-        super(
+        super(null, null,
             serverThread,
             levelStorageAccess,
             packRepository,
@@ -103,7 +_,7 @@
         LOGGER.info("Starting integrated minecraft server version {}", SharedConstants.getCurrentVersion().name());
         this.setUsesAuthentication(true);
         this.initializeKeyPair();
-        this.loadLevel();
+        this.loadLevel("");
         GameProfile host = this.getSingleplayerProfile();
         String levelName = this.getWorldData().getLevelName();
         this.setMotd(host != null ? host.name() + " - " + levelName : levelName);
