--- a/net/minecraft/client/server/IntegratedServer.java
+++ b/net/minecraft/client/server/IntegratedServer.java
@@ -16,6 +_,7 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.components.toasts.SystemToast;
 import net.minecraft.client.multiplayer.ClientPacketListener;
+import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.Services;
 import net.minecraft.server.WorldStem;
@@ -33,6 +_,7 @@
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
+import org.bukkit.command.CommandSender;
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
@@ -73,10 +_,12 @@
         this.setPvpAllowed(true);
         this.setFlightAllowed(true);
         this.initializeKeyPair();
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerAboutToStart(this);
         this.loadLevel();
         GameProfile gameprofile = this.getSingleplayerProfile();
         String s = this.getWorldData().getLevelName();
         this.setMotd(gameprofile != null ? gameprofile.getName() + " - " + s : s);
+        net.neoforged.neoforge.server.ServerLifecycleHooks.handleServerStarting(this);
         return true;
     }
 
@@ -147,6 +_,11 @@
     }
 
     @Override
+    public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+        return null;
+    }
+
+    @Override
     public Path getServerDirectory() {
         return this.minecraft.gameDirectory.toPath();
     }
@@ -225,6 +_,7 @@
 
     @Override
     public void halt(boolean p_120053_) {
+        if (isRunning())
         this.executeBlocking(() -> {
             for (ServerPlayer serverplayer : Lists.newArrayList(this.getPlayerList().getPlayers())) {
                 if (!serverplayer.getUUID().equals(this.uuid)) {
