--- a/net/minecraft/client/server/IntegratedPlayerList.java
+++ b/net/minecraft/client/server/IntegratedPlayerList.java
@@ -32,13 +_,6 @@
         super.save(p_120011_);
     }
 
-    @Override
-    public Component canPlayerLogin(SocketAddress p_120007_, GameProfile p_120008_) {
-        return (Component)(this.getServer().isSingleplayerOwner(p_120008_) && this.getPlayerByName(p_120008_.getName()) != null
-            ? Component.translatable("multiplayer.disconnect.name_taken")
-            : super.canPlayerLogin(p_120007_, p_120008_));
-    }
-
     public IntegratedServer getServer() {
         return (IntegratedServer)super.getServer();
     }
