--- a/net/minecraft/network/protocol/login/ClientboundCustomQueryPacket.java
+++ b/net/minecraft/network/protocol/login/ClientboundCustomQueryPacket.java
@@ -18,7 +_,7 @@
         this(p_179810_.readVarInt(), readPayload(p_179810_.readResourceLocation(), p_179810_));
     }
 
-    private static CustomQueryPayload readPayload(ResourceLocation p_295267_, FriendlyByteBuf p_295117_) {
+    public static CustomQueryPayload readPayload(ResourceLocation p_295267_, FriendlyByteBuf p_295117_) {
         return readUnknownPayload(p_295267_, p_295117_);
     }
 
