--- a/net/minecraft/world/level/levelgen/feature/stateproviders/DualNoiseProvider.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/DualNoiseProvider.java
@@ -14,30 +_,29 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.levelgen.LegacyRandomSource;
 import net.minecraft.world.level.levelgen.WorldgenRandom;
-import net.minecraft.world.level.levelgen.synth.Noise;
 import net.minecraft.world.level.levelgen.synth.NormalNoise;
 
 public class DualNoiseProvider extends NoiseProvider {
     public static final MapCodec<DualNoiseProvider> CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(
                 InclusiveRange.codec(Codec.INT, 1, 64).fieldOf("variety").forGetter(p -> p.variety),
-                NormalNoise.DIRECT_CODEC.fieldOf("slow_noise").forGetter(p -> p.slowNoiseParameters),
+                NormalNoise.NoiseParameters.DIRECT_CODEC.fieldOf("slow_noise").forGetter(p -> p.slowNoiseParameters),
                 ExtraCodecs.POSITIVE_FLOAT.fieldOf("slow_scale").forGetter(p -> p.slowScale)
             )
             .and(noiseProviderCodec(i))
             .apply(i, DualNoiseProvider::new)
     );
     private final InclusiveRange<Integer> variety;
-    private final NormalNoise slowNoiseParameters;
+    private final NormalNoise.NoiseParameters slowNoiseParameters;
     private final float slowScale;
-    private final Noise slowNoise;
+    private final NormalNoise slowNoise;
 
     public DualNoiseProvider(
         InclusiveRange<Integer> variety,
-        NormalNoise slowNoiseParameters,
+        NormalNoise.NoiseParameters slowNoiseParameters,
         float slowScale,
         long seed,
-        NormalNoise parameters,
+        NormalNoise.NoiseParameters parameters,
         float scale,
         List<BlockState> states
     ) {
@@ -45,7 +_,7 @@
         this.variety = variety;
         this.slowNoiseParameters = slowNoiseParameters;
         this.slowScale = slowScale;
-        this.slowNoise = slowNoiseParameters.create(new WorldgenRandom(new LegacyRandomSource(seed)));
+        this.slowNoise = NormalNoise.create(new WorldgenRandom(new LegacyRandomSource(seed)), slowNoiseParameters);
     }
 
     @Override
@@ -66,7 +_,7 @@
         return this.getRandomState(possibleStates, pos, this.scale);
     }
 
-    protected float getSlowNoiseValue(BlockPos pos) {
-        return this.slowNoise.get(pos.getX() * this.slowScale, pos.getY() * this.slowScale, pos.getZ() * this.slowScale);
+    protected double getSlowNoiseValue(BlockPos pos) {
+        return this.slowNoise.getValue(pos.getX() * this.slowScale, pos.getY() * this.slowScale, pos.getZ() * this.slowScale);
     }
 }
