--- a/net/minecraft/world/level/levelgen/feature/stateproviders/NoiseThresholdProvider.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/NoiseThresholdProvider.java
@@ -34,7 +_,7 @@
 
     public NoiseThresholdProvider(
         long seed,
-        NormalNoise parameters,
+        NormalNoise.NoiseParameters parameters,
         float scale,
         float threshold,
         float highChance,
