--- a/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
@@ -6,7 +_,6 @@
 import java.util.Collection;
 import java.util.Optional;
 import net.minecraft.core.BlockPos;
-import net.minecraft.core.Holder;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.valueproviders.IntProvider;
 import net.minecraft.util.valueproviders.IntProviders;
@@ -25,13 +_,13 @@
             )
             .apply(i, RandomizedIntStateProvider::new)
     );
-    private final Holder<BlockStateProvider> source;
+    private final BlockStateProvider source;
     private final String propertyName;
     private @Nullable IntegerProperty property;
     private final IntProvider values;
 
     public RandomizedIntStateProvider(BlockStateProvider source, IntegerProperty property, IntProvider values) {
-        this.source = Holder.direct(source);
+        this.source = source;
         this.property = property;
         this.propertyName = property.getName();
         this.values = values;
@@ -44,7 +_,7 @@
         }
     }
 
-    public RandomizedIntStateProvider(Holder<BlockStateProvider> source, String propertyName, IntProvider values) {
+    public RandomizedIntStateProvider(BlockStateProvider source, String propertyName, IntProvider values) {
         this.source = source;
         this.propertyName = propertyName;
         this.values = values;
@@ -57,7 +_,7 @@
 
     @Override
     public BlockState getState(LevelAccessor level, RandomSource random, BlockPos pos) {
-        BlockState unmodifiedState = this.source.value().getState(level, random, pos);
+        BlockState unmodifiedState = this.source.getState(level, random, pos);
         if (this.property == null || !unmodifiedState.hasProperty(this.property)) {
             IntegerProperty property = findProperty(unmodifiedState, this.propertyName);
             if (property == null) {
