--- a/net/minecraft/world/level/levelgen/feature/stateproviders/RandomBlockProvider.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/RandomBlockProvider.java
@@ -6,8 +_,8 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.Block;
@@ -16,7 +_,8 @@
 
 public record RandomBlockProvider(HolderSet<Block> blocks) implements BlockStateProvider {
     public static final MapCodec<RandomBlockProvider> CODEC = RecordCodecBuilder.mapCodec(
-        i -> i.group(RegistryCodecs.holderSet(Registries.BLOCK).fieldOf("blocks").forGetter(RandomBlockProvider::blocks)).apply(i, RandomBlockProvider::new)
+        i -> i.group(RegistryCodecs.homogeneousList(Registries.BLOCK).fieldOf("blocks").forGetter(RandomBlockProvider::blocks))
+            .apply(i, RandomBlockProvider::new)
     );
 
     @Override
