--- a/net/minecraft/world/entity/ai/gossip/GossipContainer.java
+++ b/net/minecraft/world/entity/ai/gossip/GossipContainer.java
@@ -24,6 +_,10 @@
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.VisibleForDebug;
+import net.minecraft.world.entity.npc.villager.Villager;
+import org.bukkit.craftbukkit.entity.CraftVillager;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerReputationChangeEvent;
 
 public class GossipContainer {
     public static final Codec<GossipContainer> CODEC = GossipContainer.GossipEntry.CODEC
@@ -32,6 +_,14 @@
     public static final int DISCARD_THRESHOLD = 2;
     public final Map<UUID, GossipContainer.EntityGossips> gossips = new HashMap<>();
 
+    // CraftBukkit start - store reference to villager entity
+    public Villager villager;
+
+    public GossipContainer(Villager villager) {
+        this.villager = villager;
+    }
+    // CraftBukkit end
+
     public GossipContainer() {
     }
 
@@ -106,8 +_,16 @@
     }
 
     public int getReputation(UUID entity, Predicate<GossipType> types) {
+        // CraftBukkit start - add getReputation overload with additional parameter
+        return getReputation(entity, types, true);
+    }
+
+    public int getReputation(UUID entity, Predicate<GossipType> types, boolean weighted) {
+        // CraftBukkit end
         GossipContainer.EntityGossips entry = this.gossips.get(entity);
-        return entry != null ? entry.weightedValue(types) : 0;
+        // CraftBukkit start - handle weighted parameter
+        return entry != null ? (weighted ? entry.weightedValue(types) : entry.unweightedValue(types)) : 0;
+        // CraftBukkit end
     }
 
     public long getCountForType(GossipType type, DoublePredicate valueTest) {
@@ -115,14 +_,46 @@
     }
 
     public void add(UUID target, GossipType type, int amountToAdd) {
+        // CraftBukkit start - add change reason parameter
+        add(target, type, amountToAdd, org.bukkit.entity.Villager.ReputationEvent.UNSPECIFIED);
+    }
+
+    public void add(UUID target, GossipType type, int amountToAdd, org.bukkit.entity.Villager.ReputationEvent changeReason) {
+        // CraftBukkit end
         GossipContainer.EntityGossips entityGossips = this.getOrCreate(target);
+        int oldValue = entityGossips.entries.getInt(type); // CraftBukkit - store old value
         entityGossips.entries.mergeInt(type, amountToAdd, (o, n) -> this.mergeValuesForAddition(type, o, n));
-        entityGossips.makeSureValueIsntTooLowOrTooHigh(type);
+        // CraftBukkit start - fire reputation change event
+        int newValue = entityGossips.entries.getInt(type);
+        newValue = Math.max(0, Math.min(newValue, type.max));
+        entityGossips.entries.replace(type, oldValue); // restore old value until the event completed processing
+        VillagerReputationChangeEvent event = CraftEventFactory.callVillagerReputationChangeEvent(villager, target, changeReason, CraftVillager.CraftReputationType.minecraftToBukkit(type), oldValue, newValue, type.max);
+        if (!event.isCancelled()) {
+            entityGossips.entries.replace(type, event.getNewValue());
+            entityGossips.makeSureValueIsntTooLowOrTooHigh(type);
+        }
+        // CraftBukkit end
         if (entityGossips.isEmpty()) {
             this.gossips.remove(target);
         }
     }
 
+    // CraftBukkit start
+    public void set(UUID target, GossipType type, int amount, org.bukkit.entity.Villager.ReputationEvent changeReason) {
+        int addAmount = amount - getReputation(target, Predicate.isEqual(type), false);
+        if (addAmount == 0) {
+            return;
+        }
+        this.add(target, type, addAmount, changeReason);
+    }
+    // CraftBukkit end
+
+    // CraftBukkit start - add change reason parameter
+    public void remove(UUID target, GossipType type, int amountToRemove, org.bukkit.entity.Villager.ReputationEvent changeReason) {
+        this.add(target, type, -amountToRemove, changeReason);
+    }
+    // CraftBukkit end
+
     public void remove(UUID target, GossipType type, int amountToRemove) {
         this.add(target, type, -amountToRemove);
     }
@@ -178,6 +_,16 @@
         public int weightedValue(Predicate<GossipType> types) {
             return this.entries.object2IntEntrySet().stream().filter(e -> types.test(e.getKey())).mapToInt(e -> e.getIntValue() * e.getKey().weight).sum();
         }
+
+        // CraftBukkit start
+        public int unweightedValue(Predicate<GossipType> predicate) {
+            return this.entries.object2IntEntrySet().stream().filter((entry) -> {
+                return predicate.test((GossipType) entry.getKey());
+            }).mapToInt((entry) -> {
+                return entry.getIntValue();
+            }).sum();
+        }
+        // CraftBukkit end
 
         public Stream<GossipContainer.GossipEntry> unpack(UUID target) {
             return this.entries.object2IntEntrySet().stream().map(e -> new GossipContainer.GossipEntry(target, e.getKey(), e.getIntValue()));
