--- a/net/minecraft/world/entity/ai/gossip/GossipContainer.java
+++ b/net/minecraft/world/entity/ai/gossip/GossipContainer.java
@@ -24,6 +_,9 @@
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.VisibleForDebug;
+import net.minecraft.world.entity.npc.villager.Villager;
+import org.bukkit.craftbukkit.entity.CraftVillager;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class GossipContainer {
     public static final Codec<GossipContainer> CODEC = GossipContainer.GossipEntry.CODEC
@@ -32,6 +_,14 @@
     public static final int DISCARD_THRESHOLD = 2;
     public final Map<UUID, GossipContainer.EntityGossips> gossips = new HashMap<>();
 
+    // CraftBukkit start - store reference to villager entity
+    public Villager villager;
+
+    public GossipContainer(Villager villager) {
+        this.villager = villager;
+    }
+    // CraftBukkit end
+
     public GossipContainer() {
     }
 
@@ -65,6 +_,20 @@
         return this.gossips.entrySet().stream().flatMap(e -> e.getValue().unpack(e.getKey()));
     }
 
+    // Paper start - Perf: Remove streams from hot code
+    private List<GossipContainer.GossipEntry> decompress() {
+        final List<GossipContainer.GossipEntry> entries = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+        for (final Map.Entry<UUID, GossipContainer.EntityGossips> gossips : this.gossips.entrySet()) {
+            for (final GossipContainer.GossipEntry entry : gossips.getValue().decompress(gossips.getKey())) {
+                if (entry.weightedValue() != 0) {
+                    entries.add(entry);
+                }
+            }
+        }
+        return entries;
+    }
+    // Paper end - Perf: Remove streams from hot code
+
     private Collection<GossipContainer.GossipEntry> selectGossipsForTransfer(RandomSource random, int maxCount) {
         List<GossipContainer.GossipEntry> entries = this.unpack().toList();
         if (entries.isEmpty()) {
@@ -176,8 +_,25 @@
         private final Object2IntMap<GossipType> entries = new Object2IntOpenHashMap<>();
 
         public int weightedValue(Predicate<GossipType> types) {
-            return this.entries.object2IntEntrySet().stream().filter(e -> types.test(e.getKey())).mapToInt(e -> e.getIntValue() * e.getKey().weight).sum();
-        }
+            // Paper start - Perf: Remove streams from hot code
+            int weight = 0;
+            for (Object2IntMap.Entry<GossipType> entry : this.entries.object2IntEntrySet()) {
+                if (types.test(entry.getKey())) {
+                    weight += entry.getIntValue() * entry.getKey().weight;
+                }
+            }
+            return weight;
+        }
+
+        public List<GossipContainer.GossipEntry> decompress(final UUID target) {
+            List<GossipContainer.GossipEntry> entries = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+            for (Object2IntMap.Entry<GossipType> entry : this.entries.object2IntEntrySet()) {
+                entries.add(new GossipContainer.GossipEntry(target, entry.getKey(), entry.getIntValue()));
+            }
+            return entries;
+            // Paper end - Perf: Remove streams from hot code
+        }
+
 
         public Stream<GossipContainer.GossipEntry> unpack(UUID target) {
             return this.entries.object2IntEntrySet().stream().map(e -> new GossipContainer.GossipEntry(target, e.getKey(), e.getIntValue()));
@@ -215,6 +_,44 @@
         public void remove(GossipType type) {
             this.entries.removeInt(type);
         }
+
+        // Paper start - Add villager reputation API
+        private static final GossipType[] TYPES = GossipType.values();
+
+        public com.destroystokyo.paper.entity.villager.Reputation asReputation() {
+            Map<com.destroystokyo.paper.entity.villager.ReputationType, Integer> map = new java.util.EnumMap<>(com.destroystokyo.paper.entity.villager.ReputationType.class);
+            for (Object2IntMap.Entry<GossipType> type : this.entries.object2IntEntrySet()) {
+                map.put(toApi(type.getKey()), type.getIntValue());
+            }
+
+            return new com.destroystokyo.paper.entity.villager.Reputation(map);
+        }
+
+        public void assignFromReputation(com.destroystokyo.paper.entity.villager.Reputation rep) {
+            for (GossipType type : TYPES) {
+                com.destroystokyo.paper.entity.villager.ReputationType api = toApi(type);
+
+                if (rep.hasReputationSet(api)) {
+                    int reputation = rep.getReputation(api);
+                    if (reputation == 0) {
+                        this.entries.removeInt(type);
+                    } else {
+                        this.entries.put(type, reputation);
+                    }
+                }
+            }
+        }
+
+        private static com.destroystokyo.paper.entity.villager.ReputationType toApi(GossipType type) {
+            return switch (type) {
+                case MAJOR_NEGATIVE -> com.destroystokyo.paper.entity.villager.ReputationType.MAJOR_NEGATIVE;
+                case MINOR_NEGATIVE -> com.destroystokyo.paper.entity.villager.ReputationType.MINOR_NEGATIVE;
+                case MINOR_POSITIVE -> com.destroystokyo.paper.entity.villager.ReputationType.MINOR_POSITIVE;
+                case MAJOR_POSITIVE -> com.destroystokyo.paper.entity.villager.ReputationType.MAJOR_POSITIVE;
+                case TRADING -> com.destroystokyo.paper.entity.villager.ReputationType.TRADING;
+            };
+        }
+        // Paper end - Add villager reputation API
     }
 
     private record GossipEntry(UUID target, GossipType type, int value) {
