--- a/net/minecraft/client/gui/screens/debug/GameModeSwitcherScreen.java
+++ b/net/minecraft/client/gui/screens/debug/GameModeSwitcherScreen.java
@@ -157,7 +_,7 @@
     private enum GameModeIcon {
         CREATIVE(Component.translatable("gameMode.creative"), GameType.CREATIVE, new ItemStack(Blocks.GRASS_BLOCK)),
         SURVIVAL(Component.translatable("gameMode.survival"), GameType.SURVIVAL, new ItemStack(Items.IRON_SWORD)),
-        ADVENTURE(Component.translatable("gameMode.adventure"), GameType.ADVENTURE, new ItemStack(Items.BURIED_TREASURE_MAP)),
+        ADVENTURE(Component.translatable("gameMode.adventure"), GameType.ADVENTURE, new ItemStack(Items.MAP)),
         SPECTATOR(Component.translatable("gameMode.spectator"), GameType.SPECTATOR, new ItemStack(Items.ENDER_EYE));
 
         private static final GameModeSwitcherScreen.GameModeIcon[] VALUES = values();
