--- a/net/minecraft/world/level/levelgen/structure/pools/JigsawPlacement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/JigsawPlacement.java
@@ -26,7 +_,6 @@
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.levelgen.RandomState;
 import net.minecraft.world.level.levelgen.WorldgenRandom;
-import net.minecraft.world.level.levelgen.densityfunction.SamplerContext;
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
 import net.minecraft.world.level.levelgen.structure.PoolElementStructurePiece;
 import net.minecraft.world.level.levelgen.structure.Structure;
@@ -108,18 +_,9 @@
         BoundingBox box = centerPiece.getBoundingBox();
         int centerX = (box.maxX() + box.minX()) / 2;
         int centerZ = (box.maxZ() + box.minZ()) / 2;
-        int bottomY;
-        if (projectStartToHeightmap.isEmpty()) {
-            bottomY = adjustedPosition.getY();
-        } else {
-            if (!context.couldStructureExistInColumn(centerX, centerZ, heightAccessor.getMinY(), heightAccessor.getMaxY())) {
-                return Optional.empty();
-            }
-
-            bottomY = position.getY()
-                + chunkGenerator.getFirstFreeHeight(centerX, centerZ, projectStartToHeightmap.get(), heightAccessor, context.randomState());
-        }
-
+        int bottomY = projectStartToHeightmap.isEmpty()
+            ? adjustedPosition.getY()
+            : position.getY() + chunkGenerator.getFirstFreeHeight(centerX, centerZ, projectStartToHeightmap.get(), heightAccessor, context.randomState());
         int oldAbsoluteGroundY = box.minY() + centerPiece.getGroundLevelDelta();
         centerPiece.move(0, bottomY - oldAbsoluteGroundY, 0);
         if (isStartTooCloseToWorldHeightLimits(heightAccessor, dimensionPadding, centerPiece.getBoundingBox())) {
@@ -188,7 +_,7 @@
     ) {
         for (StructureTemplate.JigsawBlockInfo jigsaw : element.getShuffledJigsawBlocks(structureTemplateManager, position, rotation, random)) {
             if (targetJigsawId.equals(jigsaw.name())) {
-                return Optional.of(jigsaw.pos());
+                return Optional.of(jigsaw.info().pos());
             }
         }
 
@@ -223,16 +_,14 @@
         ServerLevel level, Holder<StructureTemplatePool> pool, Identifier target, int maxDepth, BlockPos position, boolean keepJigsaws
     ) {
         ChunkGenerator generator = level.getChunkSource().getGenerator();
-        StructureTemplateManager structureTemplateManager = level.getStructureTemplateManager();
+        StructureTemplateManager structureTemplateManager = level.getStructureManager();
         StructureManager structureManager = level.structureManager();
         RandomSource random = level.getRandom();
-        RandomState randomState = level.getChunkSource().randomState();
         Structure.GenerationContext generationContext = new Structure.GenerationContext(
             level.registryAccess(),
             generator,
             generator.getBiomeSource(),
-            randomState.createClimateSampler(SamplerContext.EMPTY_UNCACHED),
-            randomState,
+            level.getChunkSource().randomState(),
             structureTemplateManager,
             level.getSeed(),
             ChunkPos.containing(position),
@@ -318,8 +_,9 @@
             for (StructureTemplate.JigsawBlockInfo sourceJigsaw : sourceElement.getShuffledJigsawBlocks(
                 this.structureTemplateManager, sourceBoxPosition, sourceRotation, this.random
             )) {
-                Direction sourceDirection = JigsawBlock.getFrontFacing(sourceJigsaw.state());
-                BlockPos sourceJigsawPos = sourceJigsaw.pos();
+                StructureTemplate.StructureBlockInfo sourceJigsawInfo = sourceJigsaw.info();
+                Direction sourceDirection = JigsawBlock.getFrontFacing(sourceJigsawInfo.state());
+                BlockPos sourceJigsawPos = sourceJigsawInfo.pos();
                 BlockPos targetJigsawPos = sourceJigsawPos.relative(sourceDirection);
                 int sourceJigsawLocalY = sourceJigsawPos.getY() - sourceBoxY;
                 int sourceJigsawBaseHeight = Integer.MIN_VALUE;
@@ -374,7 +_,10 @@
                                         expandTo = targetJigsaws.stream()
                                             .mapToInt(
                                                 targetJigsawx -> {
-                                                    if (!hackBox.isInside(targetJigsawx.pos().relative(JigsawBlock.getFrontFacing(targetJigsawx.state())))) {
+                                                    StructureTemplate.StructureBlockInfo targetJigsawInfo = targetJigsawx.info();
+                                                    if (!hackBox.isInside(targetJigsawInfo.pos().relative(JigsawBlock.getFrontFacing(targetJigsawInfo.state())))
+                                                        )
+                                                     {
                                                         return 0;
                                                     }
 
@@ -398,7 +_,7 @@
 
                                     for (StructureTemplate.JigsawBlockInfo targetJigsaw : targetJigsaws) {
                                         if (JigsawBlock.canAttach(sourceJigsaw, targetJigsaw)) {
-                                            BlockPos targetJigsawLocalPos = targetJigsaw.pos();
+                                            BlockPos targetJigsawLocalPos = targetJigsaw.info().pos();
                                             BlockPos rawTargetBoxPos = targetJigsawPos.subtract(targetJigsawLocalPos);
                                             BoundingBox rawTargetBB = targetElement.getBoundingBox(
                                                 this.structureTemplateManager, rawTargetBoxPos, targetRotation
@@ -407,7 +_,9 @@
                                             StructureTemplatePool.Projection targetProjection = targetElement.getProjection();
                                             boolean targetRigid = targetProjection == StructureTemplatePool.Projection.RIGID;
                                             int targetJigsawLocalY = targetJigsawLocalPos.getY();
-                                            int deltaY = sourceJigsawLocalY - targetJigsawLocalY + JigsawBlock.getFrontFacing(sourceJigsaw.state()).getStepY();
+                                            int deltaY = sourceJigsawLocalY
+                                                - targetJigsawLocalY
+                                                + JigsawBlock.getFrontFacing(sourceJigsawInfo.state()).getStepY();
                                             int targetBoxY;
                                             if (sourceRigid && targetRigid) {
                                                 targetBoxY = sourceBoxY + deltaY;
