--- a/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java
@@ -12,7 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
+import net.minecraft.resources.RegistryFileCodec;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.StringRepresentable;
 import net.minecraft.util.Util;
@@ -38,7 +_,7 @@
             .apply(i, StructureTemplatePool::new)
     );
     public static final Codec<Holder<StructureTemplatePool>> CODEC = Util.make(
-        RegistryCodecs.holder(Registries.TEMPLATE_POOL, DIRECT_CODEC), CODEC_REFERENCE::setValue
+        RegistryFileCodec.create(Registries.TEMPLATE_POOL, DIRECT_CODEC), CODEC_REFERENCE::setValue
     );
     private final List<Pair<StructurePoolElement, Integer>> rawTemplates;
     private final ObjectArrayList<StructurePoolElement> templates;
