--- a/net/minecraft/world/item/crafting/AbstractCookingRecipe.java
+++ b/net/minecraft/world/item/crafting/AbstractCookingRecipe.java
@@ -5,7 +_,7 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 
-public abstract class AbstractCookingRecipe implements Recipe<SingleRecipeInput> {
+public abstract class AbstractCookingRecipe extends io.papermc.paper.inventory.recipe.RecipeBookExactChoiceRecipe<SingleRecipeInput> implements Recipe<SingleRecipeInput> {
     protected final RecipeType<?> type;
     protected final CookingBookCategory category;
     protected final String group;
@@ -24,6 +_,7 @@
         this.result = p_252185_;
         this.experience = p_252165_;
         this.cookingTime = p_250256_;
+        this.checkExactIngredients(); // Paper - improve exact recipe choices
     }
 
     public boolean matches(SingleRecipeInput p_344849_, Level p_345973_) {
