--- a/net/minecraft/world/item/crafting/FireworkRocketRecipe.java
+++ b/net/minecraft/world/item/crafting/FireworkRocketRecipe.java
@@ -75,7 +_,7 @@
 
     @Override
     public ItemStack getResultItem(HolderLookup.Provider p_335481_) {
-        return new ItemStack(Items.FIREWORK_ROCKET);
+        return new ItemStack(Items.FIREWORK_ROCKET, 3); // Paper - Fix incorrect crafting result amount
     }
 
     @Override
