--- a/net/minecraft/world/item/crafting/ImbueRecipe.java
+++ b/net/minecraft/world/item/crafting/ImbueRecipe.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.item.crafting.display.ShapedCraftingRecipeDisplay;
 import net.minecraft.world.item.crafting.display.SlotDisplay;
 import net.minecraft.world.level.Level;
+import org.bukkit.NamespacedKey;
 
 public class ImbueRecipe extends NormalCraftingRecipe {
     public static final MapCodec<ImbueRecipe> MAP_CODEC = RecordCodecBuilder.mapCodec(
@@ -78,6 +_,17 @@
         result.set(DataComponents.POTION_CONTENTS, source.get(DataComponents.POTION_CONTENTS));
         return result;
     }
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+        org.bukkit.craftbukkit.inventory.CraftItemStack result = org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this.result.create());
+        org.bukkit.craftbukkit.inventory.CraftComplexRecipe recipe = new org.bukkit.craftbukkit.inventory.CraftComplexRecipe(id, result, this);
+        recipe.setGroup(this.group());
+        recipe.setCategory(org.bukkit.craftbukkit.inventory.CraftRecipe.getCategory(this.category()));
+        return recipe;
+    }
+    // CraftBukkit end
 
     @Override
     public RecipeSerializer<ImbueRecipe> getSerializer() {
