--- a/net/minecraft/world/item/crafting/PotionIngredient.java
+++ b/net/minecraft/world/item/crafting/PotionIngredient.java
@@ -8,7 +_,6 @@
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
-import net.minecraft.tags.ItemTags;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 
@@ -34,10 +_,5 @@
 
     public static PotionIngredient of(Item item) {
         return new PotionIngredient(Ingredient.of(item), Optional.empty());
-    }
-
-    public static boolean isPotionInput(ItemStack itemStack, RecipeAccess recipeAccess) {
-        RecipePropertySet brewingInputs = recipeAccess.propertySet(RecipePropertySet.BREWING_INPUTS);
-        return brewingInputs.test(itemStack) || itemStack.is(ItemTags.BREWING_POTION_INPUTS);
     }
 }
