--- a/net/minecraft/world/item/crafting/RecipeHolder.java
+++ b/net/minecraft/world/item/crafting/RecipeHolder.java
@@ -4,8 +_,16 @@
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.resources.ResourceKey;
+import org.bukkit.craftbukkit.util.CraftNamespacedKey;
 
 public record RecipeHolder<T extends Recipe<?>>(ResourceKey<Recipe<?>> id, T value) {
+
+    // CraftBukkit start
+    public org.bukkit.inventory.Recipe toBukkitRecipe() {
+        return this.value.toBukkitRecipe(CraftNamespacedKey.fromMinecraft(this.id.identifier()));
+    }
+    // CraftBukkit end
+
     public static final StreamCodec<RegistryFriendlyByteBuf, RecipeHolder<?>> STREAM_CODEC = StreamCodec.composite(
         ResourceKey.streamCodec(Registries.RECIPE), RecipeHolder::id, Recipe.STREAM_CODEC, RecipeHolder::value, RecipeHolder::new
     );
