--- a/net/minecraft/world/item/crafting/Recipe.java
+++ b/net/minecraft/world/item/crafting/Recipe.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.item.crafting;
 
+import com.mohistmc.youer.neoforge.ModSpecialRecipe;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
@@ -49,6 +_,10 @@
     }
 
     RecipeBookCategory recipeBookCategory();
+
+    default org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id){
+        return new ModSpecialRecipe(id, this);
+    }
 
     interface BookInfo<CategoryType> {
         CategoryType category();
