--- a/net/minecraft/world/item/crafting/RecipeType.java
+++ b/net/minecraft/world/item/crafting/RecipeType.java
@@ -12,7 +_,6 @@
     RecipeType<CampfireCookingRecipe> CAMPFIRE_COOKING = register("campfire_cooking");
     RecipeType<StonecutterRecipe> STONECUTTING = register("stonecutting");
     RecipeType<SmithingRecipe> SMITHING = register("smithing");
-    RecipeType<BrewingRecipe> BREWING = register("brewing");
 
     static <T extends Recipe<?>> RecipeType<T> register(String name) {
         return Registry.register(BuiltInRegistries.RECIPE_TYPE, Identifier.withDefaultNamespace(name), new RecipeType<T>() {
@@ -21,5 +_,15 @@
                 return name;
             }
         });
+    }
+
+    public static <T extends Recipe<?>> RecipeType<T> simple(final Identifier name) {
+        final String toString = name.toString();
+        return new RecipeType<T>() {
+            @Override
+            public String toString() {
+                return toString;
+            }
+        };
     }
 }
