--- a/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
+++ b/net/minecraft/world/item/crafting/SmithingTrimRecipe.java
@@ -17,6 +_,10 @@
 import net.minecraft.world.item.equipment.trim.ArmorTrim;
 import net.minecraft.world.item.equipment.trim.TrimMaterial;
 import net.minecraft.world.item.equipment.trim.TrimPattern;
+import org.bukkit.NamespacedKey;
+import org.bukkit.craftbukkit.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.inventory.CraftSmithingTrimRecipe;
+import org.bukkit.craftbukkit.inventory.trim.CraftTrimPattern;
 
 public class SmithingTrimRecipe extends SimpleSmithingRecipe {
     public static final MapCodec<SmithingTrimRecipe> MAP_CODEC = RecordCodecBuilder.mapCodec(
@@ -47,9 +_,18 @@
     private final Ingredient base;
     private final Ingredient addition;
     private final Holder<TrimPattern> pattern;
+    final boolean copyDataComponents; // Paper - Option to prevent data components copy
 
     public SmithingTrimRecipe(Recipe.CommonInfo commonInfo, Ingredient template, Ingredient base, Ingredient addition, Holder<TrimPattern> pattern) {
+        // Paper start - Option to prevent data components copy
+        this(commonInfo, template, base, addition, pattern, true);
+    }
+    public SmithingTrimRecipe(
+            final Recipe.CommonInfo commonInfo, final Ingredient template, final Ingredient base, final Ingredient addition, final Holder<TrimPattern> pattern, final boolean copyDataComponents
+    ) {
+        // Paper end - Option to prevent data components copy
         super(commonInfo);
+        this.copyDataComponents = copyDataComponents; // Paper
         this.template = template;
         this.base = base;
         this.addition = addition;
@@ -57,10 +_,15 @@
     }
 
     public ItemStack assemble(SmithingRecipeInput input) {
-        return applyTrim(input.base(), input.addition(), this.pattern);
+        return applyTrim(input.base(), input.addition(), this.pattern, this.copyDataComponents); // Paper start - Option to prevent data components copy
     }
 
     public static ItemStack applyTrim(ItemStack baseItem, ItemStack materialItem, Holder<TrimPattern> pattern) {
+        // Paper start - Option to prevent data components copy
+        return applyTrim(baseItem, materialItem, pattern, true);
+    }
+    public static ItemStack applyTrim(final ItemStack baseItem, final ItemStack materialItem, final Holder<TrimPattern> pattern, final boolean copyDataComponents) {
+        // Paper end - Option to prevent data components copy
         Holder<TrimMaterial> material = materialItem.get(DataComponents.PROVIDES_TRIM_MATERIAL);
         if (material != null) {
             ArmorTrim existingTrim = baseItem.get(DataComponents.TRIM);
@@ -69,7 +_,7 @@
                 return ItemStack.EMPTY;
             }
 
-            ItemStack trimmedItem = baseItem.copyWithCount(1);
+            ItemStack trimmedItem = copyDataComponents ? baseItem.copyWithCount(1) : new ItemStack(baseItem.getItem(), 1); // Paper - Option to prevent data components copy
             trimmedItem.set(DataComponents.TRIM, newTrim);
             return trimmedItem;
         } else {
@@ -117,4 +_,11 @@
             )
         );
     }
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.inventory.Recipe toBukkitRecipe(org.bukkit.NamespacedKey id) {
+        return new org.bukkit.craftbukkit.inventory.CraftSmithingTrimRecipe(id, org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(this.template), org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(this.base), org.bukkit.craftbukkit.inventory.CraftRecipe.toChoice(this.addition), org.bukkit.craftbukkit.inventory.trim.CraftTrimPattern.minecraftHolderToBukkit(this.pattern), this.copyDataComponents); // Paper - Option to prevent data components copy
+    }
+    // CraftBukkit end
 }
