--- a/com/mojang/renderpearl/backend/opengl/BufferStorage.java
+++ b/com/mojang/renderpearl/backend/opengl/BufferStorage.java
@@ -9,8 +_,8 @@
 
 @OnlyIn(Dist.CLIENT)
 public abstract class BufferStorage {
-    public static BufferStorage create(GLCapabilities capabilities, Set<String> enabledExtensions, boolean forceMutable) {
-        if (!forceMutable && capabilities.GL_ARB_buffer_storage && GlDevice.USE_GL_ARB_buffer_storage) {
+    public static BufferStorage create(GLCapabilities capabilities, Set<String> enabledExtensions) {
+        if (capabilities.GL_ARB_buffer_storage && GlDevice.USE_GL_ARB_buffer_storage) {
             enabledExtensions.add("GL_ARB_buffer_storage");
             return new BufferStorage.Immutable();
         } else {
