--- a/com/mojang/renderpearl/backend/opengl/GlBackend.java
+++ b/com/mojang/renderpearl/backend/opengl/GlBackend.java
@@ -60,17 +_,28 @@
     }
 
     @Override
-    public long createWindow(@Nullable String title, int width, int height, long flags) {
+    public void setWindowHints() {
         SDLVideo.SDL_GL_SetAttribute(17, 3);
         SDLVideo.SDL_GL_SetAttribute(18, 3);
         SDLVideo.SDL_GL_SetAttribute(20, 1);
         SDLVideo.SDL_GL_SetAttribute(19, 2);
         SDLVideo.SDL_GL_SetAttribute(22, 1);
-        return SDLVideo.SDL_CreateWindow(title, width, height, 2L | flags);
-    }
-
-    @Override
-    public GpuDevice createDevice(GpuDebugOptions debugOptions) throws BackendCreationException {
-        return new FrontendGpuDevice(new GlDevice(this, debugOptions));
+    }
+
+    @Override
+    public long getWindowFlags() {
+        return 2L;
+    }
+
+    @Override
+    public void handleWindowCreationErrors(@Nullable String error) throws BackendCreationException {
+        throw new BackendCreationException(
+            "Failed to create window with OpenGL context: " + Objects.requireNonNullElse(error, "<no error>"), BackendCreationException.Reason.OPENGL_MISSING
+        );
+    }
+
+    @Override
+    public GpuDevice createDevice(long window, GpuDebugOptions debugOptions) throws BackendCreationException {
+        return new FrontendGpuDevice(new GlDevice(window, debugOptions));
     }
 }
