--- a/com/mojang/renderpearl/backend/opengl/GlCommandEncoder.java
+++ b/com/mojang/renderpearl/backend/opengl/GlCommandEncoder.java
@@ -11,20 +_,27 @@
 import com.mojang.renderpearl.api.commands.GpuQueryPool;
 import com.mojang.renderpearl.api.commands.RenderPass;
 import com.mojang.renderpearl.api.commands.RenderPassDescriptor;
+import com.mojang.renderpearl.api.pipeline.BindGroupLayout;
+import com.mojang.renderpearl.api.pipeline.BlendFunction;
+import com.mojang.renderpearl.api.pipeline.ColorTargetState;
+import com.mojang.renderpearl.api.pipeline.DepthStencilState;
 import com.mojang.renderpearl.api.pipeline.IndexType;
+import com.mojang.renderpearl.api.pipeline.RenderPipeline;
+import com.mojang.renderpearl.api.pipeline.UniformType;
 import com.mojang.renderpearl.api.textures.GpuTexture;
 import com.mojang.renderpearl.api.textures.GpuTextureView;
 import com.mojang.renderpearl.backend.api.CommandEncoderBackend;
 import com.mojang.renderpearl.backend.api.RenderPassBackend;
-import com.mojang.renderpearl.util.TextureViewAndSampler;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.nio.ByteBuffer;
 import java.nio.IntBuffer;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 import java.util.OptionalDouble;
+import java.util.Map.Entry;
+import java.util.function.BiConsumer;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 import org.joml.Vector4fc;
@@ -38,7 +_,7 @@
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
-class GlCommandEncoder implements CommandEncoderBackend, UncheckedAutoCloseable {
+class GlCommandEncoder implements CommandEncoderBackend, AutoCloseable {
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final int MAX_SUBMITS_IN_FLIGHT = 2;
     private static final long NO_FENCE = 0L;
@@ -48,8 +_,9 @@
     private long currentSubmitIndex = 2L;
     private final int readFbo;
     private final int drawFbo;
-    private @Nullable GlRenderPipeline lastPipeline;
+    private @Nullable RenderPipeline lastPipeline;
     private @Nullable GlProgram lastProgram;
+    private VertexArrayCache.@Nullable VertexArray lastVertexArray;
     private final List<@Nullable FrameBufferAttachment> renderPassColorTextures = new ArrayList<>();
 
     protected GlCommandEncoder(GlDevice device) {
@@ -78,7 +_,7 @@
     public void submit() {
         this.fences[this.currentSubmitSlot()] = GL33C.glFenceSync(37143, 0);
         this.currentSubmitIndex++;
-        if (!this.awaitSubmit(this.currentSubmitIndex - 2L, -1L)) {
+        if (!this.awaitSubmit(this.currentSubmitIndex - 2L, Long.MAX_VALUE)) {
             throw new IllegalStateException("Failed to wait for frame completion");
         }
 
@@ -179,95 +_,64 @@
         }
 
         GlStateManager._viewport(0, 0, width, height);
-        int[] drawBuffers = new int[this.renderPassColorTextures.size()];
-
-        for (int i = 0; i < this.renderPassColorTextures.size(); i++) {
-            if (colorAttachments.get(i) != null) {
-                drawBuffers[i] = 36064 + i;
-            } else {
-                drawBuffers[i] = 0;
-            }
-        }
-
-        GL33C.glDrawBuffers(drawBuffers);
         this.lastPipeline = null;
         ScissorState scissorState = new ScissorState();
         scissorState.enable(renderArea.x(), renderArea.y(), renderArea.width(), renderArea.height());
-        return new GlRenderPass(this, this.device, this.renderPassColorTextures.size(), scissorState);
+        return new GlRenderPass(this, this.device, depthAttachment != null, this.renderPassColorTextures.size(), scissorState);
     }
 
     @Override
     public void clearColorTexture(GpuTexture colorTexture, Vector4fc clearColor) {
+        this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, ((GlTexture)colorTexture).id, 0, 0, 36160);
         GL33C.glClearColor(clearColor.x(), clearColor.y(), clearColor.z(), clearColor.w());
         GlStateManager._disableScissorTest();
         GlStateManager._colorMask(15);
-
-        for (int i = 0; i < colorTexture.getMipLevels(); i++) {
-            this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, ((GlTexture)colorTexture).id, 0, i, 36160);
-            GlStateManager._clear(16384);
-        }
-
+        GlStateManager._clear(16384);
         GlStateManager._glFramebufferTexture2D(36160, 36064, 3553, 0, 0);
         GlStateManager._glBindFramebuffer(36160, 0);
     }
 
     @Override
     public void clearColorAndDepthTextures(GpuTexture colorTexture, Vector4fc clearColor, GpuTexture depthTexture, double clearDepth) {
+        int fbo = this.device
+            .frameBufferCache()
+            .getFbo(this.device.directStateAccess(), Collections.singletonList((GlTexture)colorTexture), (GlTexture)depthTexture);
+        GlStateManager._glBindFramebuffer(36160, fbo);
         GlStateManager._disableScissorTest();
         GL33C.glClearDepth(clearDepth);
         GL33C.glClearColor(clearColor.x(), clearColor.y(), clearColor.z(), clearColor.w());
         GlStateManager._depthMask(true);
         GlStateManager._colorMask(15);
-
-        for (int i = 0; i < colorTexture.getMipLevels(); i++) {
-            int fbo = this.device
-                .frameBufferCache()
-                .getFbo(this.device.directStateAccess(), Collections.singletonList((GlTexture)colorTexture), (GlTexture)depthTexture, i);
-            GlStateManager._glBindFramebuffer(36160, fbo);
-            GlStateManager._clear(16640);
-        }
-
+        GlStateManager._clear(16640);
         GlStateManager._glBindFramebuffer(36160, 0);
     }
 
     @Override
     public void clearColorAndDepthTextures(
-        GpuTexture colorTexture,
-        Vector4fc clearColor,
-        GpuTexture depthTexture,
-        double clearDepth,
-        int regionX,
-        int regionY,
-        int regionWidth,
-        int regionHeight,
-        int mipLevel
+        GpuTexture colorTexture, Vector4fc clearColor, GpuTexture depthTexture, double clearDepth, int regionX, int regionY, int regionWidth, int regionHeight
     ) {
+        int fbo = this.device
+            .frameBufferCache()
+            .getFbo(this.device.directStateAccess(), Collections.singletonList((GlTexture)colorTexture), (GlTexture)depthTexture);
+        GlStateManager._glBindFramebuffer(36160, fbo);
         GlStateManager._scissorBox(regionX, regionY, regionWidth, regionHeight);
         GlStateManager._enableScissorTest();
         GL33C.glClearDepth(clearDepth);
         GL33C.glClearColor(clearColor.x(), clearColor.y(), clearColor.z(), clearColor.w());
         GlStateManager._depthMask(true);
         GlStateManager._colorMask(15);
-        int fbo = this.device
-            .frameBufferCache()
-            .getFbo(this.device.directStateAccess(), Collections.singletonList((GlTexture)colorTexture), (GlTexture)depthTexture, mipLevel);
-        GlStateManager._glBindFramebuffer(36160, fbo);
         GlStateManager._clear(16640);
         GlStateManager._glBindFramebuffer(36160, 0);
     }
 
     @Override
     public void clearDepthTexture(GpuTexture depthTexture, double clearDepth) {
+        this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, 0, ((GlTexture)depthTexture).id, 0, 36160);
+        GL33C.glDrawBuffer(0);
         GL33C.glClearDepth(clearDepth);
         GlStateManager._depthMask(true);
         GlStateManager._disableScissorTest();
-
-        for (int i = 0; i < depthTexture.getMipLevels(); i++) {
-            this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, 0, ((GlTexture)depthTexture).id, i, 36160);
-            GL33C.glDrawBuffer(0);
-            GlStateManager._clear(256);
-        }
-
+        GlStateManager._clear(256);
         GL33C.glDrawBuffer(36064);
         GlStateManager._glFramebufferTexture2D(36160, 36096, 3553, 0, 0);
         GlStateManager._glBindFramebuffer(36160, 0);
@@ -303,10 +_,8 @@
         GlStateManager._pixelStore(3314, width);
         GlStateManager._pixelStore(3316, 0);
         GlStateManager._pixelStore(3315, 0);
-        GlStateManager._pixelStore(3317, destination.getFormat().byteAlignment());
-        GlStateManager._texSubImage2D(
-            target, mipLevel, destX, destY, width, height, GlConst.toGlExternalId(destination.getFormat()), GlConst.toGlType(destination.getFormat()), source
-        );
+        GlStateManager._pixelStore(3317, destination.getFormat().componentCount());
+        GlStateManager._texSubImage2D(target, mipLevel, destX, destY, width, height, GlConst.toGlExternalId(destination.getFormat()), 5121, source);
     }
 
     @Override
@@ -366,19 +_,12 @@
     public void copyTextureToBuffer(GpuTexture source, GpuBuffer destination, long offset, Runnable callback, int mipLevel, int x, int y, int width, int height) {
         ((GlBuffer)destination).checkCanBeUsed();
         GlStateManager.clearGlErrors();
-        boolean isDepth = source.getFormat().hasDepthAspect();
-        int textureId = ((GlTexture)source).glId();
-        this.device.directStateAccess().bindFrameBufferTextures(this.readFbo, !isDepth ? textureId : 0, isDepth ? textureId : 0, mipLevel, 36008);
+        this.device.directStateAccess().bindFrameBufferTextures(this.readFbo, ((GlTexture)source).glId(), 0, mipLevel, 36008);
         GlStateManager._glBindBuffer(35051, ((GlBuffer)destination).handle());
-        GlStateManager._pixelStore(3333, source.getFormat().byteAlignment());
         GlStateManager._pixelStore(3330, width);
-        if (isDepth) {
-            GlStateManager._glReadBuffer(0);
-        }
-
         GlStateManager._readPixels(x, y, width, height, GlConst.toGlExternalId(source.getFormat()), GlConst.toGlType(source.getFormat()), offset);
         RenderSystem.queueFencedTask(callback);
-        GlStateManager._glFramebufferTexture2D(36008, isDepth ? 36096 : 36064, 3553, 0, mipLevel);
+        GlStateManager._glFramebufferTexture2D(36008, 36064, 3553, 0, mipLevel);
         GlStateManager._glBindFramebuffer(36008, 0);
         GlStateManager._glBindBuffer(35051, 0);
         int error = GlStateManager._getError();
@@ -396,8 +_,8 @@
         boolean isDepth = source.getFormat().hasDepthAspect();
         int sourceId = ((GlTexture)source).glId();
         int destId = ((GlTexture)destination).glId();
-        this.device.directStateAccess().bindFrameBufferTextures(this.readFbo, isDepth ? 0 : sourceId, isDepth ? sourceId : 0, mipLevel, 0);
-        this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, isDepth ? 0 : destId, isDepth ? destId : 0, mipLevel, 0);
+        this.device.directStateAccess().bindFrameBufferTextures(this.readFbo, isDepth ? 0 : sourceId, isDepth ? sourceId : 0, 0, 0);
+        this.device.directStateAccess().bindFrameBufferTextures(this.drawFbo, isDepth ? 0 : destId, isDepth ? destId : 0, 0, 0);
         this.device
             .directStateAccess()
             .blitFrameBuffers(this.readFbo, this.drawFbo, sourceX, sourceY, width, height, destX, destY, width, height, isDepth ? 256 : 16384, 9728);
@@ -409,8 +_,7 @@
         }
     }
 
-    void presentTexture(long windowHandle, GpuTextureView textureView, int swapchainWidth, int swapchainHeight) {
-        this.device.makeCurrent(windowHandle);
+    public void presentTexture(GpuTextureView textureView, int swapchainWidth, int swapchainHeight) {
         int destY = Math.max(0, swapchainHeight - textureView.getHeight(0));
         int copyWidth = Math.min(swapchainWidth, textureView.getWidth(0));
         int copyHeight = Math.min(swapchainHeight, textureView.getHeight(0));
@@ -427,30 +_,115 @@
         return new GlFence(this);
     }
 
+    protected <T> void executeDrawMultiple(
+        GlRenderPass renderPass,
+        Collection<RenderPass.Draw<T>> draws,
+        @Nullable GpuBuffer defaultIndexBuffer,
+        @Nullable IndexType defaultIndexType,
+        Collection<String> dynamicUniforms,
+        T uniformArgument
+    ) {
+        if (this.trySetup(renderPass, dynamicUniforms)) {
+            if (defaultIndexType == null) {
+                defaultIndexType = IndexType.SHORT;
+            }
+
+            for (RenderPass.Draw<T> draw : draws) {
+                IndexType indexType = draw.indexType() == null ? defaultIndexType : draw.indexType();
+                renderPass.setIndexBuffer(draw.indexBuffer() == null ? defaultIndexBuffer : draw.indexBuffer(), indexType);
+                renderPass.setVertexBuffer(draw.slot(), draw.vertexBuffer().slice());
+                if (GlRenderPass.VALIDATION) {
+                    if (renderPass.indexBuffer == null) {
+                        throw new IllegalStateException("Missing index buffer");
+                    }
+
+                    ((GlBuffer)renderPass.indexBuffer).checkCanBeUsed();
+                    if (renderPass.indexBuffer.isClosed()) {
+                        throw new IllegalStateException("Index buffer has been closed!");
+                    }
+
+                    if (draw.slot() < 0 || draw.slot() >= 16) {
+                        throw new IllegalStateException("Vertex buffer slot must be between 0 and 16");
+                    }
+
+                    if (renderPass.vertexBuffers[draw.slot()] != null) {
+                        ((GlBuffer)renderPass.vertexBuffers[draw.slot()].buffer()).checkCanBeUsed();
+                    }
+
+                    if (renderPass.vertexBuffers[draw.slot()] == null) {
+                        throw new IllegalStateException("Missing vertex buffer at slot " + draw.slot());
+                    }
+
+                    if (renderPass.vertexBuffers[draw.slot()].buffer().isClosed()) {
+                        throw new IllegalStateException("Vertex buffer at slot " + draw.slot() + " has been closed!");
+                    }
+                }
+
+                BiConsumer<T, RenderPass.UniformUploader> uniformUploaderConsumer = draw.uniformUploaderConsumer();
+                if (uniformUploaderConsumer != null) {
+                    uniformUploaderConsumer.accept(uniformArgument, (name, buffer) -> {
+                        ((GlBuffer)buffer.buffer()).checkCanBeUsed();
+                        if (renderPass.pipeline.program().getUniform(name) instanceof Uniform.Ubo(int blockBinding)) {
+                            GL33C.glBindBufferRange(35345, blockBinding, ((GlBuffer)buffer.buffer()).handle(), buffer.offset(), buffer.length());
+                        }
+                    });
+                }
+
+                if (renderPass.vertexBufferDirty) {
+                    this.lastVertexArray = this.device
+                        .vertexArrayCache()
+                        .bindVertexArray(renderPass.pipeline.info().getVertexFormatBindings(), renderPass.vertexBuffers, null);
+                    renderPass.vertexBufferDirty = false;
+                }
+
+                this.drawFromBuffers(renderPass, draw.baseVertex(), draw.firstIndex(), draw.indexCount(), indexType, renderPass.pipeline, 1, 0);
+            }
+        }
+    }
+
+    private void validateDraw(GlRenderPass renderPass, @Nullable IndexType indexType) {
+        if (GlRenderPass.VALIDATION) {
+            if (indexType != null) {
+                if (renderPass.indexBuffer == null) {
+                    throw new IllegalStateException("Missing index buffer");
+                }
+
+                ((GlBuffer)renderPass.indexBuffer).checkCanBeUsed();
+                if (renderPass.indexBuffer.isClosed()) {
+                    throw new IllegalStateException("Index buffer has been closed!");
+                }
+
+                if ((renderPass.indexBuffer.usage() & 64) == 0) {
+                    throw new IllegalStateException("Index buffer must have GpuBuffer.USAGE_INDEX!");
+                }
+            }
+
+            GlRenderPipeline pipeline = renderPass.pipeline;
+
+            for (int i = 0; i < 16; i++) {
+                if (renderPass.vertexBuffers[i] == null && pipeline != null && pipeline.info().getVertexFormatBindings()[i] != null) {
+                    throw new IllegalStateException("Vertex format contains elements but vertex buffer at slot " + i + " is null");
+                }
+
+                if (renderPass.vertexBuffers[i] != null) {
+                    ((GlBuffer)renderPass.vertexBuffers[i].buffer()).checkCanBeUsed();
+                }
+            }
+        }
+    }
+
     protected void executeDraw(
         GlRenderPass renderPass, int baseVertex, int firstIndex, int drawCount, @Nullable IndexType indexType, int instanceCount, int firstInstance
     ) {
-        this.setupDraw(renderPass);
-        if (indexType != null) {
-            if (firstInstance > 0) {
-                ARBBaseInstance.glDrawElementsInstancedBaseVertexBaseInstance(
-                    renderPass.pipeline.primitiveTopology(),
-                    drawCount,
-                    GlConst.toGl(indexType),
-                    (long)firstIndex * indexType.bytes,
-                    instanceCount,
-                    baseVertex,
-                    firstInstance
-                );
-            } else {
-                GL33C.glDrawElementsInstancedBaseVertex(
-                    renderPass.pipeline.primitiveTopology(), drawCount, GlConst.toGl(indexType), (long)firstIndex * indexType.bytes, instanceCount, baseVertex
-                );
-            }
-        } else if (firstInstance > 0) {
-            ARBBaseInstance.glDrawArraysInstancedBaseInstance(renderPass.pipeline.primitiveTopology(), baseVertex, drawCount, instanceCount, firstInstance);
-        } else {
-            GL33C.glDrawArraysInstanced(renderPass.pipeline.primitiveTopology(), baseVertex, drawCount, instanceCount);
+        if (this.trySetup(renderPass, Collections.emptyList())) {
+            this.validateDraw(renderPass, indexType);
+            this.lastVertexArray = this.device
+                .vertexArrayCache()
+                .bindVertexArray(
+                    renderPass.pipeline.info().getVertexFormatBindings(), renderPass.vertexBuffers, renderPass.vertexBufferDirty ? null : this.lastVertexArray
+                );
+            renderPass.vertexBufferDirty = false;
+            this.drawFromBuffers(renderPass, baseVertex, firstIndex, drawCount, indexType, renderPass.pipeline, instanceCount, firstInstance);
         }
     }
 
@@ -462,126 +_,310 @@
         IntBuffer vertexOffsets,
         int drawCount
     ) {
-        this.setupDraw(renderPass);
-        if (indexType == null) {
-            GL33C.nglMultiDrawArrays(
-                renderPass.pipeline.primitiveTopology(), MemoryUtil.memAddress(vertexOffsets), MemoryUtil.memAddress(indexCounts), drawCount
-            );
-        } else {
-            assert firstIndexOffsets != null;
-            GL33C.nglMultiDrawElementsBaseVertex(
-                renderPass.pipeline.primitiveTopology(),
-                MemoryUtil.memAddress(indexCounts),
-                GlConst.toGl(indexType),
-                MemoryUtil.memAddress(firstIndexOffsets),
-                drawCount,
-                MemoryUtil.memAddress(vertexOffsets)
-            );
+        if (this.trySetup(renderPass, Collections.emptyList())) {
+            this.validateDraw(renderPass, indexType);
+            this.lastVertexArray = this.device
+                .vertexArrayCache()
+                .bindVertexArray(
+                    renderPass.pipeline.info().getVertexFormatBindings(), renderPass.vertexBuffers, renderPass.vertexBufferDirty ? null : this.lastVertexArray
+                );
+            renderPass.vertexBufferDirty = false;
+            if (indexType == null) {
+                GL33C.nglMultiDrawArrays(
+                    GlConst.toGl(renderPass.pipeline.info().getPrimitiveTopology()),
+                    MemoryUtil.memAddress(vertexOffsets),
+                    MemoryUtil.memAddress(indexCounts),
+                    drawCount
+                );
+            } else {
+                GlStateManager._glBindBuffer(34963, ((GlBuffer)renderPass.indexBuffer).handle());
+                assert firstIndexOffsets != null;
+                GL33C.nglMultiDrawElementsBaseVertex(
+                    GlConst.toGl(renderPass.pipeline.info().getPrimitiveTopology()),
+                    MemoryUtil.memAddress(indexCounts),
+                    GlConst.toGl(indexType),
+                    MemoryUtil.memAddress(firstIndexOffsets),
+                    drawCount,
+                    MemoryUtil.memAddress(vertexOffsets)
+                );
+            }
         }
     }
 
     protected void executeDrawIndirect(GlRenderPass renderPass, @Nullable IndexType indexType, GlBuffer commands, long offset, int drawCount) {
-        this.setupDraw(renderPass);
-        GlStateManager._glBindBuffer(36671, commands.handle());
-        if (indexType == null) {
-            if (drawCount > 1) {
-                ARBMultiDrawIndirect.glMultiDrawArraysIndirect(renderPass.pipeline.primitiveTopology(), offset, drawCount, 0);
+        if (this.trySetup(renderPass, Collections.emptyList())) {
+            this.validateDraw(renderPass, indexType);
+            this.lastVertexArray = this.device
+                .vertexArrayCache()
+                .bindVertexArray(
+                    renderPass.pipeline.info().getVertexFormatBindings(), renderPass.vertexBuffers, renderPass.vertexBufferDirty ? null : this.lastVertexArray
+                );
+            renderPass.vertexBufferDirty = false;
+            GlStateManager._glBindBuffer(36671, commands.handle());
+            if (indexType == null) {
+                if (drawCount > 1) {
+                    ARBMultiDrawIndirect.glMultiDrawArraysIndirect(GlConst.toGl(renderPass.pipeline.info().getPrimitiveTopology()), offset, drawCount, 0);
+                } else {
+                    ARBDrawIndirect.glDrawArraysIndirect(GlConst.toGl(renderPass.pipeline.info().getPrimitiveTopology()), offset);
+                }
             } else {
-                ARBDrawIndirect.glDrawArraysIndirect(renderPass.pipeline.primitiveTopology(), offset);
+                GlStateManager._glBindBuffer(34963, ((GlBuffer)renderPass.indexBuffer).handle());
+                if (drawCount > 1) {
+                    ARBMultiDrawIndirect.glMultiDrawElementsIndirect(
+                        GlConst.toGl(renderPass.pipeline.info().getPrimitiveTopology()), GlConst.toGl(indexType), offset, drawCount, 0
+                    );
+                } else {
+                    ARBDrawIndirect.glDrawElementsIndirect(GlConst.toGl(renderPass.pipeline.info().getPrimitiveTopology()), GlConst.toGl(indexType), offset);
+                }
             }
-        } else if (drawCount > 1) {
-            ARBMultiDrawIndirect.glMultiDrawElementsIndirect(renderPass.pipeline.primitiveTopology(), GlConst.toGl(indexType), offset, drawCount, 0);
-        } else {
-            ARBDrawIndirect.glDrawElementsIndirect(renderPass.pipeline.primitiveTopology(), GlConst.toGl(indexType), offset);
         }
     }
 
-    private void setupDraw(GlRenderPass renderPass) {
-        if (renderPass.vertexBufferDirty) {
-            renderPass.pipeline.vertexArray().bind(renderPass.vertexBuffers);
-            renderPass.vertexBufferDirty = false;
-        }
-
-        if (renderPass.indexBufferDirty) {
-            renderPass.indexBufferDirty = false;
+    private void drawFromBuffers(
+        GlRenderPass renderPass,
+        int baseVertex,
+        int firstIndex,
+        int drawCount,
+        @Nullable IndexType indexType,
+        GlRenderPipeline pipeline,
+        int instanceCount,
+        int firstInstance
+    ) {
+        if (indexType != null) {
             GlStateManager._glBindBuffer(34963, ((GlBuffer)renderPass.indexBuffer).handle());
-        }
-
-        if (this.lastPipeline != renderPass.pipeline) {
-            renderPass.pipeline.bind();
-            this.lastPipeline = renderPass.pipeline;
-        }
-
+            if (firstInstance > 0) {
+                ARBBaseInstance.glDrawElementsInstancedBaseVertexBaseInstance(
+                    GlConst.toGl(pipeline.info().getPrimitiveTopology()),
+                    drawCount,
+                    GlConst.toGl(indexType),
+                    (long)firstIndex * indexType.bytes,
+                    instanceCount,
+                    baseVertex,
+                    firstInstance
+                );
+            } else {
+                GL33C.glDrawElementsInstancedBaseVertex(
+                    GlConst.toGl(pipeline.info().getPrimitiveTopology()),
+                    drawCount,
+                    GlConst.toGl(indexType),
+                    (long)firstIndex * indexType.bytes,
+                    instanceCount,
+                    baseVertex
+                );
+            }
+        } else if (firstInstance > 0) {
+            ARBBaseInstance.glDrawArraysInstancedBaseInstance(
+                GlConst.toGl(pipeline.info().getPrimitiveTopology()), baseVertex, drawCount, instanceCount, firstInstance
+            );
+        } else {
+            GL33C.glDrawArraysInstanced(GlConst.toGl(pipeline.info().getPrimitiveTopology()), baseVertex, drawCount, instanceCount);
+        }
+    }
+
+    private boolean trySetup(GlRenderPass renderPass, Collection<String> dynamicUniforms) {
+        if (GlRenderPass.VALIDATION) {
+            if (renderPass.pipeline == null) {
+                throw new IllegalStateException("Can't draw without a render pipeline");
+            }
+
+            for (BindGroupLayout.UniformDescription uniform : BindGroupLayout.flattenUniforms(renderPass.pipeline.info().getBindGroupLayouts())) {
+                GpuBufferSlice value = renderPass.uniforms.get(uniform.name());
+                if (!dynamicUniforms.contains(uniform.name())) {
+                    if (value == null) {
+                        throw new IllegalStateException("Missing uniform " + uniform.name() + " (should be " + uniform.type() + ")");
+                    }
+
+                    ((GlBuffer)value.buffer()).checkCanBeUsed();
+                    if (uniform.type() == UniformType.UNIFORM_BUFFER) {
+                        if (value.buffer().isClosed()) {
+                            throw new IllegalStateException("Uniform buffer " + uniform.name() + " is already closed");
+                        }
+
+                        if ((value.buffer().usage() & 128) == 0) {
+                            throw new IllegalStateException("Uniform buffer " + uniform.name() + " must have GpuBuffer.USAGE_UNIFORM");
+                        }
+                    }
+
+                    if (uniform.type() == UniformType.TEXEL_BUFFER) {
+                        if (value.offset() != 0L || value.length() != value.buffer().size()) {
+                            throw new IllegalStateException("Uniform texel buffers do not support a slice of a buffer, must be entire buffer");
+                        }
+
+                        if ((value.buffer().usage() & 256) == 0) {
+                            throw new IllegalStateException("Uniform texel buffer " + uniform.name() + " must have GpuBuffer.USAGE_UNIFORM_TEXEL_BUFFER");
+                        }
+
+                        if (uniform.gpuFormat() == null) {
+                            throw new IllegalStateException("Invalid uniform texel buffer " + uniform.name() + " (missing a texture format)");
+                        }
+                    }
+                }
+            }
+
+            for (Entry<String, Uniform> entry : renderPass.pipeline.program().getUniforms().entrySet()) {
+                if (entry.getValue() instanceof Uniform.Sampler) {
+                    String name = entry.getKey();
+                    GlRenderPass.TextureViewAndSampler viewAndSampler = renderPass.samplers.get(name);
+                    if (viewAndSampler == null) {
+                        throw new IllegalStateException("Missing sampler " + name);
+                    }
+
+                    GlTextureView textureView = viewAndSampler.view();
+                    if (textureView.isClosed()) {
+                        throw new IllegalStateException("Texture view " + name + " (" + textureView.texture().getLabel() + ") has been closed!");
+                    }
+
+                    if ((textureView.texture().usage() & 4) == 0) {
+                        throw new IllegalStateException("Texture view " + name + " (" + textureView.texture().getLabel() + ") must have USAGE_TEXTURE_BINDING!");
+                    }
+
+                    if (viewAndSampler.sampler().isClosed()) {
+                        throw new IllegalStateException("Sampler for " + name + " (" + textureView.texture().getLabel() + ") has been closed!");
+                    }
+                }
+            }
+
+            if (renderPass.pipeline.info().wantsDepthTexture() && !renderPass.hasDepthTexture()) {
+                LOGGER.warn("Render pipeline {} wants a depth texture but none was provided - this is probably a bug", renderPass.pipeline.info().getLocation());
+            }
+        } else if (renderPass.pipeline == null) {
+            return false;
+        }
+
+        RenderPipeline pipeline = renderPass.pipeline.info();
         GlProgram glProgram = renderPass.pipeline.program();
-        if (renderPass.pushConstantsDirty) {
-            renderPass.pushConstantsDirty = false;
-            Uniform.Ubo pushConstantUBO = glProgram.pushConstant();
-            if (pushConstantUBO != null) {
-                assert renderPass.pushConstants != null;
-                GL33C.glBindBufferRange(
-                    35345,
-                    pushConstantUBO.blockBinding(),
-                    ((GlBuffer)renderPass.pushConstants.buffer()).handle(),
-                    renderPass.pushConstants.offset(),
-                    renderPass.pushConstants.length()
-                );
-            }
-        }
-
-        if (renderPass.anyUniformDirty) {
-            renderPass.anyUniformDirty = false;
-
-            for (int i = 0; i < renderPass.dirtyUniforms.size(); i++) {
-                if (renderPass.dirtyUniforms.getBoolean(i)) {
-                    renderPass.dirtyUniforms.set(i, false);
-                    Uniform dirtyUniform = glProgram.getUniform(i);
-                    if (dirtyUniform != null) {
-                        switch (dirtyUniform) {
-                            case Uniform.Ubo(int blockBinding):
-                                GpuBufferSlice var26 = (GpuBufferSlice)renderPass.uniforms.get(i);
-                                GL33C.glBindBufferRange(35345, blockBinding, ((GlBuffer)var26.buffer()).handle(), var26.offset(), var26.length());
-                                break;
-                            case Uniform.Utb(int samplerIndex, GpuFormat format, int texture):
-                                GlStateManager._activeTexture(33984 + samplerIndex);
-                                GL33C.glBindTexture(35882, texture);
-                                GpuBufferSlice var30 = (GpuBufferSlice)renderPass.uniforms.get(i);
-                                GL33C.glTexBuffer(35882, GlConst.toGlInternalId(format), ((GlBuffer)var30.buffer()).handle());
-                                break;
-                            case Uniform.Sampler(int samplerIndex):
-                                TextureViewAndSampler viewAndSampler = (TextureViewAndSampler)renderPass.uniforms.get(i);
-                                if (viewAndSampler != null) {
-                                    GlTextureView textureView = (GlTextureView)viewAndSampler.view();
-                                    GlStateManager._activeTexture(33984 + samplerIndex);
-                                    GlTexture texture = textureView.texture();
-                                    int target;
-                                    if ((texture.usage() & 16) != 0) {
-                                        target = 34067;
-                                        GL33C.glBindTexture(34067, texture.id);
-                                    } else {
-                                        target = 3553;
-                                        GlStateManager._bindTexture(texture.id);
-                                    }
-
-                                    GL33C.glBindSampler(samplerIndex, ((GlSampler)viewAndSampler.sampler()).getId());
-                                    GlStateManager._texParameter(target, 33084, textureView.baseMipLevel());
-                                    GlStateManager._texParameter(target, 33085, textureView.baseMipLevel() + textureView.mipLevels() - 1);
-                                }
-                                break;
-                            default:
-                                throw new MatchException(null, null);
-                        }
-                    }
-                }
-            }
-        }
-
-        if (renderPass.scissorStateDirty) {
-            renderPass.scissorStateDirty = false;
-            if (renderPass.isScissorEnabled()) {
-                GlStateManager._enableScissorTest();
-                GlStateManager._scissorBox(renderPass.getScissorX(), renderPass.getScissorY(), renderPass.getScissorWidth(), renderPass.getScissorHeight());
-            } else {
-                GlStateManager._disableScissorTest();
+        this.applyPipelineState(pipeline);
+        boolean differentProgram = this.lastProgram != glProgram;
+        if (differentProgram) {
+            GlStateManager._glUseProgram(glProgram.getProgramId());
+            this.lastProgram = glProgram;
+        }
+
+        for (Entry<String, Uniform> entry : glProgram.getUniforms().entrySet()) {
+            String name = entry.getKey();
+            boolean isDirty = renderPass.dirtyUniforms.contains(name);
+            switch ((Uniform)entry.getValue()) {
+                case Uniform.Ubo(int blockBinding):
+                    if (isDirty) {
+                        GpuBufferSlice bufferView = renderPass.uniforms.get(name);
+                        GL33C.glBindBufferRange(35345, blockBinding, ((GlBuffer)bufferView.buffer()).handle(), bufferView.offset(), bufferView.length());
+                    }
+                    break;
+                case Uniform.Utb(int location, int samplerIndex, GpuFormat format, int texture):
+                    if (differentProgram || isDirty) {
+                        GlStateManager._glUniform1i(location, samplerIndex);
+                    }
+
+                    GlStateManager._activeTexture(33984 + samplerIndex);
+                    GL33C.glBindTexture(35882, texture);
+                    if (isDirty) {
+                        GpuBufferSlice bufferView = renderPass.uniforms.get(name);
+                        GL33C.glTexBuffer(35882, GlConst.toGlInternalId(format), ((GlBuffer)bufferView.buffer()).handle());
+                    }
+                    break;
+                case Uniform.Sampler(int location, int samplerIndex):
+                    GlRenderPass.TextureViewAndSampler viewAndSampler = renderPass.samplers.get(name);
+                    if (viewAndSampler == null) {
+                        break;
+                    }
+
+                    GlTextureView textureView = viewAndSampler.view();
+                    if (differentProgram || isDirty) {
+                        GlStateManager._glUniform1i(location, samplerIndex);
+                    }
+
+                    GlStateManager._activeTexture(33984 + samplerIndex);
+                    GlTexture texture = textureView.texture();
+                    int target;
+                    if ((texture.usage() & 16) != 0) {
+                        target = 34067;
+                        GL33C.glBindTexture(34067, texture.id);
+                    } else {
+                        target = 3553;
+                        GlStateManager._bindTexture(texture.id);
+                    }
+
+                    GL33C.glBindSampler(samplerIndex, viewAndSampler.sampler().getId());
+                    GlStateManager._texParameter(target, 33084, textureView.baseMipLevel());
+                    GlStateManager._texParameter(target, 33085, textureView.baseMipLevel() + textureView.mipLevels() - 1);
+                    break;
+                default:
+                    throw new MatchException(null, null);
+            }
+        }
+
+        renderPass.dirtyUniforms.clear();
+        int[] drawBuffers = new int[renderPass.colorAttachmentCount];
+
+        for (int i = 0; i < renderPass.colorAttachmentCount; i++) {
+            drawBuffers[i] = 36064 + i;
+        }
+
+        GL33C.glDrawBuffers(drawBuffers);
+        if (renderPass.isScissorEnabled()) {
+            GlStateManager._enableScissorTest();
+            GlStateManager._scissorBox(renderPass.getScissorX(), renderPass.getScissorY(), renderPass.getScissorWidth(), renderPass.getScissorHeight());
+        } else {
+            GlStateManager._disableScissorTest();
+        }
+
+        return true;
+    }
+
+    private void applyPipelineState(RenderPipeline pipeline) {
+        if (this.lastPipeline != pipeline) {
+            this.lastPipeline = pipeline;
+            DepthStencilState depthStencilState = pipeline.getDepthStencilState();
+            if (depthStencilState != null) {
+                GlStateManager._enableDepthTest();
+                GlStateManager._depthFunc(GlConst.toGl(depthStencilState.depthTest()));
+                GlStateManager._depthMask(depthStencilState.writeDepth());
+                if (depthStencilState.depthBiasConstant() == 0.0F && depthStencilState.depthBiasScaleFactor() == 0.0F) {
+                    GlStateManager._disablePolygonOffset();
+                } else {
+                    GlStateManager._polygonOffset(depthStencilState.depthBiasScaleFactor(), depthStencilState.depthBiasConstant());
+                    GlStateManager._enablePolygonOffset();
+                }
+            } else {
+                GlStateManager._disableDepthTest();
+                GlStateManager._depthMask(false);
+                GlStateManager._disablePolygonOffset();
+            }
+
+            if (pipeline.isCull()) {
+                GlStateManager._enableCull();
+            } else {
+                GlStateManager._disableCull();
+            }
+
+            ColorTargetState[] colorTargetStates = pipeline.getColorTargetStates();
+
+            for (int i = 0; i < colorTargetStates.length; i++) {
+                ColorTargetState state = colorTargetStates[i];
+                if (state != null) {
+                    if (state.blendFunction().isPresent()) {
+                        GlStateManager._enableBlend(i);
+                        BlendFunction blendFunction = state.blendFunction().get();
+                        GlStateManager._blendFuncSeparate(
+                            GlConst.toGl(blendFunction.color().sourceFactor()),
+                            GlConst.toGl(blendFunction.color().destFactor()),
+                            GlConst.toGl(blendFunction.alpha().sourceFactor()),
+                            GlConst.toGl(blendFunction.alpha().destFactor())
+                        );
+                        GlStateManager._blendEquationSeparate(GlConst.toGl(blendFunction.color().op()), GlConst.toGl(blendFunction.alpha().op()));
+                    } else {
+                        GlStateManager._disableBlend(i);
+                    }
+                }
+            }
+
+            GlStateManager._polygonMode(1032, GlConst.toGl(pipeline.getPolygonMode()));
+
+            for (int i = 0; i < colorTargetStates.length; i++) {
+                ColorTargetState state = colorTargetStates[i];
+                if (state != null) {
+                    GlStateManager._colorMask(i, state.writeMask());
+                }
             }
         }
     }
