--- a/com/mojang/renderpearl/backend/opengl/GlDevice.java
+++ b/com/mojang/renderpearl/backend/opengl/GlDevice.java
@@ -1,5 +_,6 @@
 package com.mojang.renderpearl.backend.opengl;
 
+import com.mojang.blaze3d.preprocessor.GlslPreprocessor;
 import com.mojang.logging.LogUtils;
 import com.mojang.renderpearl.api.GpuFormat;
 import com.mojang.renderpearl.api.buffers.GpuBuffer;
@@ -8,30 +_,32 @@
 import com.mojang.renderpearl.api.device.DeviceInfo;
 import com.mojang.renderpearl.api.device.GpuDebugOptions;
 import com.mojang.renderpearl.api.device.GpuOutOfMemoryException;
+import com.mojang.renderpearl.api.pipeline.RenderPipeline;
+import com.mojang.renderpearl.api.pipeline.ShaderSource;
+import com.mojang.renderpearl.api.pipeline.ShaderType;
 import com.mojang.renderpearl.api.textures.AddressMode;
 import com.mojang.renderpearl.api.textures.FilterMode;
 import com.mojang.renderpearl.api.textures.GpuSampler;
 import com.mojang.renderpearl.api.textures.GpuTexture;
 import com.mojang.renderpearl.api.textures.GpuTextureView;
-import com.mojang.renderpearl.backend.api.BackendRenderPipeline;
 import com.mojang.renderpearl.backend.api.CommandEncoderBackend;
 import com.mojang.renderpearl.backend.api.GpuDeviceBackend;
 import com.mojang.renderpearl.backend.api.GpuSurfaceBackend;
 import java.nio.ByteBuffer;
-import java.nio.IntBuffer;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
-import java.util.Map;
 import java.util.Objects;
 import java.util.OptionalDouble;
 import java.util.Set;
-import java.util.function.BiFunction;
-import java.util.function.BooleanSupplier;
 import java.util.function.Supplier;
+import net.minecraft.client.renderer.ShaderDefines;
+import net.minecraft.client.renderer.ShaderManager;
+import net.minecraft.resources.Identifier;
 import net.minecraft.util.Mth;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
+import org.apache.commons.lang3.StringUtils;
 import org.jspecify.annotations.Nullable;
 import org.lwjgl.opengl.ARBClipControl;
 import org.lwjgl.opengl.GL;
@@ -39,7 +_,6 @@
 import org.lwjgl.opengl.GLCapabilities;
 import org.lwjgl.sdl.SDLError;
 import org.lwjgl.sdl.SDLVideo;
-import org.lwjgl.system.MemoryStack;
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
@@ -55,62 +_,33 @@
     protected static boolean USE_GL_ARB_draw_indirect = true;
     protected static boolean USE_GL_ARB_multi_draw_indirect = true;
     protected static boolean USE_GL_ARB_shader_draw_parameters = true;
-    private final long initialWindowHandle;
+    private final long windowHandle;
     private final long glContext;
     private final GlCommandEncoder encoder;
     private final @Nullable GlDebug debugLog;
     private final GlDebugLabel debugLabels;
     private final DirectStateAccess directStateAccess;
     private final FrameBufferCache frameBufferCache = new FrameBufferCache();
-    private final BiFunction<GlProgram, BackendRenderPipeline.CreateInfo, VertexArray> vertexArraySource;
+    private final VertexArrayCache vertexArrayCache;
     private final BufferStorage bufferStorage;
     private final DeviceInfo deviceInfo;
-    private final GlPipelineRecompiler recompiler;
+    private final ShaderDefines globalDefines;
     private boolean shaderCompilerRequiresSacrifice = true;
-    private long currentWindow = 0L;
-
-    public GlDevice(GlBackend backend, GpuDebugOptions debugOptions) throws BackendCreationException {
-        this.initialWindowHandle = backend.createWindow("Minecraft - RenderPearl OpenGL Hidden Utility Window", 320, 480, 2147614728L);
-        if (this.initialWindowHandle == 0L) {
-            throw new BackendCreationException(
-                "Failed to create window for OpenGL context: " + Objects.requireNonNullElse(SDLError.SDL_GetError(), "<no error>"),
-                BackendCreationException.Reason.OPENGL_MISSING
-            );
-        }
-
-        long glContext = SDLVideo.SDL_GL_CreateContext(this.initialWindowHandle);
+
+    public GlDevice(long windowHandle, GpuDebugOptions debugOptions) throws BackendCreationException {
+        this.windowHandle = windowHandle;
+        long glContext = SDLVideo.SDL_GL_CreateContext(windowHandle);
         if (glContext == 0L) {
-            SDLVideo.SDL_DestroyWindow(this.initialWindowHandle);
             throw new BackendCreationException(
-                "Failed to create OpenGL context: " + Objects.requireNonNullElse(SDLError.SDL_GetError(), "<no error>"),
+                "Failed to create window with OpenGL context: " + Objects.requireNonNullElse(SDLError.SDL_GetError(), "<no error>"),
                 BackendCreationException.Reason.OPENGL_MISSING
             );
         }
 
         this.glContext = glContext;
-        this.makeCurrent(this.initialWindowHandle);
-
-        try (MemoryStack stack = MemoryStack.stackPush()) {
-            IntBuffer majorVersion = stack.callocInt(1);
-            IntBuffer minorVersion = stack.callocInt(1);
-            SDLVideo.SDL_GL_GetAttribute(17, majorVersion);
-            SDLVideo.SDL_GL_GetAttribute(18, minorVersion);
-            if (majorVersion.get(0) < 3 || majorVersion.get(0) == 3 && minorVersion.get(0) < 3) {
-                throw new BackendCreationException(
-                    "Failed to create OpenGL 3.3 context, got OpenGL " + majorVersion.get(0) + "." + minorVersion.get(0),
-                    BackendCreationException.Reason.OPENGL_MISSING
-                );
-            }
-
-            long secondWindow = backend.createWindow("Minecraft - RenderPearl OpenGL Hidden Test Window", 320, 480, 2147614728L);
-            if (secondWindow == 0L) {
-                throw new BackendCreationException(
-                    "Failed to create window for OpenGL after creating context: " + Objects.requireNonNullElse(SDLError.SDL_GetError(), "<no error>"),
-                    BackendCreationException.Reason.OPENGL_MISSING
-                );
-            }
-
-            SDLVideo.SDL_DestroyWindow(secondWindow);
+        SDLVideo.SDL_GL_MakeCurrent(windowHandle, glContext);
+
+        try {
             GLCapabilities capabilities = GL.createCapabilities();
             Set<String> enabledExtensions = new HashSet<>();
             int maxSupportedAnisotropy;
@@ -121,12 +_,12 @@
                 maxSupportedAnisotropy = 1;
             }
 
-            GlHeuristics heuristics = new GlHeuristics(GlStateManager._getString(7937), GlStateManager._getString(7936));
+            GlHeuristics heuristics = new GlHeuristics(GlStateManager._getString(7937));
             this.debugLog = GlDebug.enableDebugCallback(debugOptions.logLevel(), debugOptions.synchronousLogs(), enabledExtensions);
             this.debugLabels = GlDebugLabel.create(capabilities, debugOptions.useLabels(), enabledExtensions);
-            this.bufferStorage = BufferStorage.create(capabilities, enabledExtensions, heuristics.couldBeIntelGen7() || heuristics.isNvidia());
+            this.vertexArrayCache = VertexArrayCache.create(capabilities, this.debugLabels, enabledExtensions);
+            this.bufferStorage = BufferStorage.create(capabilities, enabledExtensions);
             this.directStateAccess = DirectStateAccess.create(capabilities, enabledExtensions, heuristics);
-            this.vertexArraySource = VertexArray.createSource(capabilities, enabledExtensions);
             GL33C.glEnable(34895);
             GL33C.glEnable(34370);
             if (capabilities.GL_ARB_clip_control) {
@@ -151,10 +_,14 @@
 
             this.deviceInfo = heuristics.createDeviceInfo(capabilities, maxSupportedAnisotropy, enabledExtensions);
             this.encoder = new GlCommandEncoder(this);
-            this.recompiler = new GlPipelineRecompiler(this.debugLabels, this.deviceInfo.features().shaderDrawParameters());
+            ShaderDefines.Builder globalDefinesBuilder = ShaderDefines.builder();
+            if (this.deviceInfo.isZZeroToOne()) {
+                globalDefinesBuilder.define("B3D_DEPTH_IS_ZERO_TO_ONE");
+            }
+
+            this.globalDefines = globalDefinesBuilder.build();
         } catch (Throwable throwable) {
             SDLVideo.SDL_GL_DestroyContext(glContext);
-            SDLVideo.SDL_DestroyWindow(this.initialWindowHandle);
             throw throwable;
         }
     }
@@ -164,8 +_,8 @@
     }
 
     @Override
-    public GpuSurfaceBackend createSurface(long windowHandle, BooleanSupplier isIconified) {
-        return new GlSurface(windowHandle, isIconified);
+    public GpuSurfaceBackend createSurface(long windowHandle) {
+        return new GlSurface(windowHandle);
     }
 
     @Override
@@ -182,6 +_,13 @@
 
     @Override
     public GpuTexture createTexture(
+        @Nullable Supplier<String> label, @GpuTexture.Usage int usage, GpuFormat format, int width, int height, int depthOrLayers, int mipLevels
+    ) {
+        return this.createTexture(this.debugLabels.exists() && label != null ? label.get() : null, usage, format, width, height, depthOrLayers, mipLevels);
+    }
+
+    @Override
+    public GpuTexture createTexture(
         @Nullable String label, @GpuTexture.Usage int usage, GpuFormat format, int width, int height, int depthOrLayers, int mipLevels
     ) {
         GlStateManager.clearGlErrors();
@@ -241,6 +_,11 @@
     }
 
     @Override
+    public GpuTextureView createTextureView(GpuTexture texture) {
+        return this.createTextureView(texture, 0, texture.getMipLevels());
+    }
+
+    @Override
     public GpuTextureView createTextureView(GpuTexture texture, int baseMipLevel, int mipLevels) {
         return new GlTextureView((GlTexture)texture, baseMipLevel, mipLevels, this.frameBufferCache);
     }
@@ -311,27 +_,69 @@
     @Override
     public void close() {
         this.encoder.close();
+        SDLVideo.SDL_GL_MakeCurrent(this.windowHandle, this.glContext);
         SDLVideo.SDL_GL_DestroyContext(this.glContext);
-        SDLVideo.SDL_DestroyWindow(this.initialWindowHandle);
     }
 
     public DirectStateAccess directStateAccess() {
         return this.directStateAccess;
     }
 
-    @Override
-    public BackendRenderPipeline.Pending compilePipeline(BackendRenderPipeline.CreateInfo createInfo) {
-        Map<BackendRenderPipeline.CreateInfo.Shader, String> decompiledShaders = this.recompiler.decompileShaders(createInfo);
-        return decompiledShaders == null ? BackendRenderPipeline.Pending.NULL : () -> {
-            this.sacrificeShaderToOpenGlAndAmd();
-            GlProgram glProgram = this.recompiler.compileProgram(createInfo, decompiledShaders);
-            if (glProgram == null) {
-                return null;
+    protected GlShaderModule compileShader(Identifier id, ShaderType type, ShaderDefines defines, ShaderSource shaderSource) {
+        String source = shaderSource.get(id, type);
+        if (source == null) {
+            LOGGER.error("Couldn't find source for {} shader ({})", type, id);
+            return GlShaderModule.INVALID_SHADER;
+        } else {
+            String sourceWithDefines = GlslPreprocessor.injectDefines(source, defines);
+            sourceWithDefines = GlslPreprocessor.injectDefines(sourceWithDefines, this.globalDefines);
+            int shaderId = GlStateManager.glCreateShader(GlConst.toGl(type));
+            GlStateManager.glShaderSource(shaderId, sourceWithDefines);
+            GlStateManager.glCompileShader(shaderId);
+            if (GlStateManager.glGetShaderi(shaderId, 35713) == 0) {
+                String logInfo = StringUtils.trim(GlStateManager.glGetShaderInfoLog(shaderId, 32768));
+                LOGGER.error("Couldn't compile {} shader ({}): {}", type.getName(), id, logInfo);
+                return GlShaderModule.INVALID_SHADER;
+            } else {
+                GlShaderModule module = new GlShaderModule(shaderId, id, type);
+                this.debugLabels.applyLabel(module);
+                return module;
             }
-
-            VertexArray vertexArray = this.vertexArraySource.apply(glProgram, createInfo);
-            return new GlRenderPipeline(this, createInfo, glProgram, vertexArray);
-        };
+        }
+    }
+
+    private @Nullable GlProgram compileProgram(RenderPipeline pipeline, ShaderSource shaderSource) {
+        GlShaderModule vertexShader = this.compileShader(pipeline.getVertexShader(), ShaderType.VERTEX, pipeline.getShaderDefines(), shaderSource);
+        GlShaderModule fragmentShader = this.compileShader(pipeline.getFragmentShader(), ShaderType.FRAGMENT, pipeline.getShaderDefines(), shaderSource);
+        if (vertexShader == GlShaderModule.INVALID_SHADER) {
+            LOGGER.error("Couldn't compile pipeline {}: vertex shader {} was invalid", pipeline.getLocation(), pipeline.getVertexShader());
+            return null;
+        }
+
+        if (fragmentShader == GlShaderModule.INVALID_SHADER) {
+            LOGGER.error("Couldn't compile pipeline {}: fragment shader {} was invalid", pipeline.getLocation(), pipeline.getFragmentShader());
+            return null;
+        }
+
+        try {
+            GlProgram compiled = GlProgram.link(vertexShader, fragmentShader, pipeline.getVertexFormatBindings(), pipeline.getLocation().toString());
+            compiled.setupBindGroupLayouts(pipeline.getBindGroupLayouts());
+            this.debugLabels.applyLabel(compiled);
+            return compiled;
+        } catch (IllegalArgumentException | ShaderManager.CompilationException e) {
+            LOGGER.error("Couldn't compile program for pipeline {}", pipeline.getLocation(), e);
+            return null;
+        }
+    }
+
+    public @Nullable GlRenderPipeline compilePipeline(RenderPipeline pipeline, ShaderSource shaderSource) {
+        this.sacrificeShaderToOpenGlAndAmd();
+        GlProgram glProgram = this.compileProgram(pipeline, shaderSource);
+        return glProgram == null ? null : new GlRenderPipeline(this, pipeline, glProgram);
+    }
+
+    public VertexArrayCache vertexArrayCache() {
+        return this.vertexArrayCache;
     }
 
     public BufferStorage getBufferStorage() {
@@ -348,21 +_,12 @@
     }
 
     @Override
-    public long getTimestampCalibrationOffset() {
-        long deviceTime = GL33C.glGetInteger64(36392);
-        long hostTime = System.nanoTime();
-        return hostTime - deviceTime;
+    public long getTimestampNow() {
+        return GL33C.glGetInteger64(36392);
     }
 
     @Override
     public DeviceInfo getDeviceInfo() {
         return this.deviceInfo;
-    }
-
-    void makeCurrent(long windowHandle) {
-        if (windowHandle != this.currentWindow) {
-            this.currentWindow = windowHandle;
-            SDLVideo.SDL_GL_MakeCurrent(windowHandle, this.glContext);
-        }
     }
 }
