--- a/com/mojang/renderpearl/backend/opengl/GlHeuristics.java
+++ b/com/mojang/renderpearl/backend/opengl/GlHeuristics.java
@@ -24,14 +_,10 @@
     private static final List<String> DEVICE_NAMES_THAT_IMPLY_VIRTUAL = List.of("virtgl");
     private final boolean isGlOnDx12;
     private final boolean isAmd;
-    private final boolean isNvidia;
-    private final boolean couldBeIntelGen7;
 
-    GlHeuristics(String deviceName, String vendor) {
+    GlHeuristics(String deviceName) {
         this.isGlOnDx12 = isGlOnDx12(deviceName);
         this.isAmd = isAmd(deviceName);
-        this.isNvidia = isNvidia(deviceName);
-        this.couldBeIntelGen7 = couldBeIntelGen7(deviceName.toLowerCase(Locale.ROOT), vendor.toLowerCase(Locale.ROOT));
     }
 
     public boolean isGlOnDx12() {
@@ -42,14 +_,6 @@
         return this.isAmd;
     }
 
-    public boolean isNvidia() {
-        return this.isNvidia;
-    }
-
-    public boolean couldBeIntelGen7() {
-        return this.couldBeIntelGen7;
-    }
-
     private static boolean isGlOnDx12(String deviceName) {
         boolean isWindowsArm64 = Util.getPlatform() == Util.OS.WINDOWS && Util.isAarch64();
         return isWindowsArm64 || deviceName.startsWith("D3D12");
@@ -59,10 +_,6 @@
         return deviceName.contains("AMD");
     }
 
-    private static boolean isNvidia(String deviceName) {
-        return deviceName.toLowerCase(Locale.ROOT).contains("nvidia");
-    }
-
     private static int getMaxSupportedTextureSize() {
         int maxReported = GlStateManager._getInteger(3379);
 
@@ -82,11 +_,6 @@
     public DeviceInfo createDeviceInfo(GLCapabilities capabilities, int maxSupportedAnisotropy, Set<String> enabledExtensions) {
         String renderer = GlStateManager._getString(7937);
         String vendor = GlStateManager._getString(7936);
-        String rendererLowerCase = renderer.toLowerCase(Locale.ROOT);
-        String vendorLowerCase = vendor.toLowerCase(Locale.ROOT);
-        int drawIndirectCount = enabledExtensions.contains("GL_ARB_multi_draw_indirect")
-            ? Integer.MAX_VALUE
-            : (enabledExtensions.contains("GL_ARB_draw_indirect") ? 1 : 0);
         return new DeviceInfo(
             renderer,
             vendor,
@@ -94,17 +_,8 @@
             capabilities.GL_ARB_clip_control,
             "OpenGL",
             1.0F,
-            new DeviceLimits(
-                maxSupportedAnisotropy,
-                GL33C.glGetInteger(35380),
-                getMaxSupportedTextureSize(),
-                Long.MAX_VALUE,
-                0,
-                GL33C.glGetInteger(34852),
-                drawIndirectCount
-            ),
+            new DeviceLimits(maxSupportedAnisotropy, GL33C.glGetInteger(35380), getMaxSupportedTextureSize(), Long.MAX_VALUE, 0, GL33C.glGetInteger(34852)),
             new DeviceFeatures(
-                true,
                 enabledExtensions.contains("GL_ARB_shader_draw_parameters"),
                 false,
                 true,
@@ -114,23 +_,9 @@
                 enabledExtensions.contains("GL_ARB_buffer_storage")
             ),
             Collections.unmodifiableSet(enabledExtensions),
-            new HintsAndWorkarounds(
-                this.isGlOnDx12(), this.isAmd(), Util.isAppleSiliconMac(renderer), vendorLowerCase.contains("intel") && !rendererLowerCase.contains("arc")
-            ),
-            this.guessDeviceType(rendererLowerCase, vendorLowerCase)
+            new HintsAndWorkarounds(this.isGlOnDx12(), this.isAmd()),
+            this.guessDeviceType(renderer.toLowerCase(Locale.ROOT), vendor.toLowerCase(Locale.ROOT))
         );
-    }
-
-    private static boolean couldBeIntelGen7(String renderer, String vendor) {
-        if (!vendor.contains("intel")) {
-            return false;
-        } else if (renderer.contains("2500")) {
-            return true;
-        } else if (renderer.contains("4000")) {
-            return true;
-        } else {
-            return renderer.contains("hd graphics (byt)") ? true : renderer.endsWith("hd graphics");
-        }
     }
 
     private DeviceType guessDeviceType(String renderer, String vendor) {
