--- a/com/mojang/renderpearl/backend/opengl/GlShaderModule.java
+++ b/com/mojang/renderpearl/backend/opengl/GlShaderModule.java
@@ -2,20 +_,20 @@
 
 import com.mojang.blaze3d.systems.RenderSystem;
 import com.mojang.renderpearl.api.pipeline.ShaderType;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
+import net.minecraft.resources.Identifier;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public class GlShaderModule implements UncheckedAutoCloseable {
+public class GlShaderModule implements AutoCloseable {
     private static final int NOT_ALLOCATED = -1;
-    public static final GlShaderModule INVALID_SHADER = new GlShaderModule(-1, "invalid", ShaderType.VERTEX);
-    private final String label;
+    public static final GlShaderModule INVALID_SHADER = new GlShaderModule(-1, Identifier.withDefaultNamespace("invalid"), ShaderType.VERTEX);
+    private final Identifier id;
     private int shaderId;
     private final ShaderType type;
 
-    public GlShaderModule(int shaderId, String label, ShaderType type) {
-        this.label = label;
+    public GlShaderModule(int shaderId, Identifier id, ShaderType type) {
+        this.id = id;
         this.shaderId = shaderId;
         this.type = type;
     }
@@ -31,15 +_,15 @@
         this.shaderId = -1;
     }
 
-    public String getLabel() {
-        return this.label;
+    public Identifier getId() {
+        return this.id;
     }
 
     public int getShaderId() {
         return this.shaderId;
     }
 
-    public ShaderType getType() {
-        return this.type;
+    public String getDebugLabel() {
+        return this.type.idConverter().idToFile(this.id).toString();
     }
 }
