--- a/com/mojang/renderpearl/backend/opengl/GlStateManager.java
+++ b/com/mojang/renderpearl/backend/opengl/GlStateManager.java
@@ -25,8 +_,7 @@
     private static final Plot PLOT_BUFFERS = TracyClient.createPlot("GPU Buffers");
     private static int numBuffers = 0;
     private static final boolean IS_MACOS = Util.getPlatform() == Util.OS.OSX;
-    private static final GlStateManager.BlendState BLEND = new GlStateManager.BlendState();
-    private static final boolean[] BLEND_ENABLE = new boolean[8];
+    private static final GlStateManager.BlendState[] BLEND = new GlStateManager.BlendState[8];
     private static final GlStateManager.DepthState DEPTH = new GlStateManager.DepthState();
     private static final GlStateManager.CullState CULL = new GlStateManager.CullState();
     private static final GlStateManager.PolygonOffsetState POLY_OFFSET = new GlStateManager.PolygonOffsetState();
@@ -84,36 +_,32 @@
 
     public static void _disableBlend(int index) {
         RenderSystem.assertOnRenderThread();
-        if (BLEND_ENABLE[index]) {
-            BLEND_ENABLE[index] = false;
-            GL33C.glDisablei(3042, index);
-        }
+        BLEND[index].mode.disable();
     }
 
     public static void _enableBlend(int index) {
         RenderSystem.assertOnRenderThread();
-        if (!BLEND_ENABLE[index]) {
-            BLEND_ENABLE[index] = true;
-            GL33C.glEnablei(3042, index);
-        }
+        BLEND[index].mode.enable();
     }
 
     public static void _blendFuncSeparate(int srcRgb, int dstRgb, int srcAlpha, int dstAlpha) {
         RenderSystem.assertOnRenderThread();
-        if (srcRgb != BLEND.srcRgb || dstRgb != BLEND.dstRgb || srcAlpha != BLEND.srcAlpha || dstAlpha != BLEND.dstAlpha) {
-            BLEND.srcRgb = srcRgb;
-            BLEND.dstRgb = dstRgb;
-            BLEND.srcAlpha = srcAlpha;
-            BLEND.dstAlpha = dstAlpha;
+        GlStateManager.BlendState firstBlend = BLEND[0];
+        if (srcRgb != firstBlend.srcRgb || dstRgb != firstBlend.dstRgb || srcAlpha != firstBlend.srcAlpha || dstAlpha != firstBlend.dstAlpha) {
+            firstBlend.srcRgb = srcRgb;
+            firstBlend.dstRgb = dstRgb;
+            firstBlend.srcAlpha = srcAlpha;
+            firstBlend.dstAlpha = dstAlpha;
             glBlendFuncSeparate(srcRgb, dstRgb, srcAlpha, dstAlpha);
         }
     }
 
     public static void _blendEquationSeparate(int modeRgb, int modeAlpha) {
         RenderSystem.assertOnRenderThread();
-        if (modeRgb != BLEND.modeRgb || modeAlpha != BLEND.modeAlpha) {
-            BLEND.modeRgb = modeRgb;
-            BLEND.modeAlpha = modeAlpha;
+        GlStateManager.BlendState firstBlend = BLEND[0];
+        if (modeRgb != firstBlend.modeRgb || modeAlpha != firstBlend.modeAlpha) {
+            firstBlend.modeRgb = modeRgb;
+            firstBlend.modeAlpha = modeAlpha;
             glBlendEquationSeparate(modeRgb, modeAlpha);
         }
     }
@@ -305,11 +_,6 @@
         GL33C.glFramebufferTexture2D(target, attachment, textarget, texture, level);
     }
 
-    public static void _glReadBuffer(int mode) {
-        RenderSystem.assertOnRenderThread();
-        GL33C.glReadBuffer(mode);
-    }
-
     public static void glBlendFuncSeparate(int srcColor, int dstColor, int srcAlpha, int dstAlpha) {
         RenderSystem.assertOnRenderThread();
         GL33C.glBlendFuncSeparate(srcColor, dstColor, srcAlpha, dstAlpha);
@@ -566,10 +_,11 @@
 
     static {
         Arrays.setAll(COLOR_MASK, var0 -> 15);
-        Arrays.fill(BLEND_ENABLE, false);
+        Arrays.setAll(BLEND, var0 -> new GlStateManager.BlendState());
     }
 
     private static class BlendState {
+        public final GlStateManager.BooleanState mode = new GlStateManager.BooleanState(3042);
         public int srcRgb = 1;
         public int dstRgb = 0;
         public int modeRgb = 32774;
