--- a/com/mojang/renderpearl/backend/opengl/GlSurface.java
+++ b/com/mojang/renderpearl/backend/opengl/GlSurface.java
@@ -8,7 +_,6 @@
 import java.util.Collection;
 import java.util.EnumSet;
 import java.util.Set;
-import java.util.function.BooleanSupplier;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 import org.lwjgl.sdl.SDLVideo;
@@ -17,13 +_,11 @@
 public class GlSurface implements GpuSurfaceBackend {
     private static final Set<GpuSurface.PresentMode> SUPPORTED_PRESENT_MODES = EnumSet.of(GpuSurface.PresentMode.FIFO, GpuSurface.PresentMode.IMMEDIATE);
     private final long windowHandle;
-    private final BooleanSupplier isIconified;
     private int swapchainWidth;
     private int swapchainHeight;
 
-    public GlSurface(long windowHandle, BooleanSupplier isIconified) {
+    public GlSurface(long windowHandle) {
         this.windowHandle = windowHandle;
-        this.isIconified = isIconified;
     }
 
     @Override
@@ -39,15 +_,12 @@
     }
 
     @Override
-    public void acquireNextTexture() throws SurfaceException {
-        if (this.isIconified.getAsBoolean()) {
-            throw new SurfaceException("Cannot acquire minimized window");
-        }
+    public void acquireNextTexture() {
     }
 
     @Override
     public void blitFromTexture(CommandEncoderBackend commandEncoder, GpuTextureView textureView) {
-        ((GlCommandEncoder)commandEncoder).presentTexture(this.windowHandle, textureView, this.swapchainWidth, this.swapchainHeight);
+        ((GlCommandEncoder)commandEncoder).presentTexture(textureView, this.swapchainWidth, this.swapchainHeight);
     }
 
     @Override
