--- a/com/mojang/renderpearl/backend/opengl/GlTransientMemory.java
+++ b/com/mojang/renderpearl/backend/opengl/GlTransientMemory.java
@@ -4,7 +_,6 @@
 import com.mojang.renderpearl.api.buffers.GpuBufferSlice;
 import com.mojang.renderpearl.api.buffers.TransientMemory;
 import com.mojang.renderpearl.backend.util.TransientBlockAllocator;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 import it.unimi.dsi.fastutil.ints.IntComparator;
 import it.unimi.dsi.fastutil.objects.ReferenceArrayList;
@@ -20,7 +_,7 @@
 import org.lwjgl.system.MemoryUtil;
 
 @OnlyIn(Dist.CLIENT)
-public abstract class GlTransientMemory implements TransientMemory, UncheckedAutoCloseable {
+public abstract class GlTransientMemory implements TransientMemory, AutoCloseable {
     private static final long BLOCK_SIZE = 524288L;
     private static final long MAX_CPU_ALIGNMENT = 16L;
     private static final long MAX_GPU_ALIGNMENT = Long.highestOneBit(Long.MAX_VALUE);
@@ -28,8 +_,8 @@
     protected final DirectStateAccess dsa;
     protected final BufferStorage bufferStorage;
     protected final GlDebugLabel debugLabels;
-    private final TransientBlockAllocator<TransientBlockAllocator.Allocator.CpuBlock> cpuBlockAllocator = new TransientBlockAllocator<>(
-        524288L, 16L, TransientBlockAllocator.Allocator.CpuBlock.memalloc()
+    private final TransientBlockAllocator<Long> cpuBlockAllocator = new TransientBlockAllocator<>(
+        524288L, 16L, TransientBlockAllocator.Allocator.create(MemoryUtil::nmemAlloc, MemoryUtil::nmemFree)
     );
 
     GlTransientMemory(GlDevice device, GlCommandEncoder encoder) {
@@ -46,9 +_,8 @@
     @Override
     public ByteBuffer allocateCpu(long size, long alignment, long minimumAllocation, long elementSize) {
         assert size <= 2147483647L;
-        TransientBlockAllocator.Allocation<TransientBlockAllocator.Allocator.CpuBlock> alloc = this.cpuBlockAllocator
-            .allocate(size, alignment, minimumAllocation, elementSize);
-        return MemoryUtil.memByteBuffer(alloc.block().address() + alloc.offset(), (int)alloc.size());
+        TransientBlockAllocator.Allocation<Long> alloc = this.cpuBlockAllocator.allocate(size, alignment, minimumAllocation, elementSize);
+        return MemoryUtil.memByteBuffer(alloc.block() + alloc.offset(), (int)alloc.size());
     }
 
     public void rotate() {
@@ -87,7 +_,7 @@
         }
 
         @Override
-        public GpuBufferSlice.MappedView allocateStaging(long size, long alignment, @GpuBuffer.Usage int usage, long minimumAllocation, long elementSize) {
+        public GpuBufferSlice.MappedView allocateStaging(long size, long alignment, int usage, long minimumAllocation, long elementSize) {
             TransientBlockAllocator.Allocation<GlTransientMemory.Fallback.GlAllocation> allocation = this.blockAllocator
                 .allocate(size, alignment, minimumAllocation, elementSize);
             GlTransientMemory.TransientGpuBuffer transientBuffer = new GlTransientMemory.TransientGpuBuffer(
@@ -109,7 +_,7 @@
         }
 
         @Override
-        public GpuBufferSlice allocateGpu(long size, long alignment, @GpuBuffer.Usage int usage, long minimumAllocation, long elementSize) {
+        public GpuBufferSlice allocateGpu(long size, long alignment, int usage, long minimumAllocation, long elementSize) {
             TransientBlockAllocator.Allocation<GlTransientMemory.Fallback.GlAllocation> allocation = this.blockAllocator
                 .allocate(size, alignment, minimumAllocation, elementSize);
             GlTransientMemory.TransientGpuBuffer transientBuffer = new GlTransientMemory.TransientGpuBuffer(
@@ -119,7 +_,7 @@
         }
 
         @Override
-        public GpuBufferSlice.MappedView allocateGpuMapped(long size, long alignment, @GpuBuffer.Usage int usage, long minimumAllocation, long elementSize) {
+        public GpuBufferSlice.MappedView allocateGpuMapped(long size, long alignment, int usage, long minimumAllocation, long elementSize) {
             return this.allocateStaging(size, alignment, usage, minimumAllocation, elementSize);
         }
 
@@ -129,7 +_,7 @@
         }
 
         @Override
-        public GpuBufferSlice uploadGpu(List<ByteBuffer> data, long alignment, @GpuBuffer.Usage int usage, long minimumAllocation, long elementSize) {
+        public GpuBufferSlice uploadGpu(List<ByteBuffer> data, long alignment, int usage, long minimumAllocation, long elementSize) {
             long totalSize = 0L;
 
             for (int i = 0; i < data.size(); i++) {
@@ -161,7 +_,7 @@
         }
 
         @Override
-        public List<GpuBufferSlice> multiUploadGpu(List<ByteBuffer> data, long alignment, @GpuBuffer.Usage int usage) {
+        public List<GpuBufferSlice> multiUploadGpu(List<ByteBuffer> data, long alignment, int usage) {
             ReferenceArrayList<GpuBufferSlice> uploadedBuffers = new ReferenceArrayList<>();
             uploadedBuffers.size(data.size());
             IntArrayList sortedDataIndices = IntArrayList.toList(IntStream.range(0, data.size()));
@@ -203,11 +_,7 @@
             return uploadedBuffers;
         }
 
-        private record GlAllocation(GlBuffer glBuffer, long hostBuffer) implements TransientBlockAllocator.Allocator.Block {
-            @Override
-            public boolean suboptimal() {
-                return false;
-            }
+        private record GlAllocation(GlBuffer glBuffer, long hostBuffer) {
         }
     }
 
@@ -250,13 +_,13 @@
         @Override
         public void rotate() {
             GlTransientMemory.PersistentMapping.Rotation previousRotation = this.rotations[this.encoder.currentSubmitSlot()];
-            this.rotations[this.encoder.currentSubmitSlot()] = new GlTransientMemory.PersistentMapping.Rotation(
-                this.stagingBlockAllocator.rotate(), this.gpuBlockAllocator.rotate(), this.gpuMappedBlockAllocator.rotate()
-            );
             if (previousRotation != null) {
                 previousRotation.run();
             }
 
+            this.rotations[this.encoder.currentSubmitSlot()] = new GlTransientMemory.PersistentMapping.Rotation(
+                this.stagingBlockAllocator.rotate(), this.gpuBlockAllocator.rotate(), this.gpuMappedBlockAllocator.rotate()
+            );
             super.rotate();
         }
 
@@ -302,7 +_,7 @@
         }
 
         @Override
-        public GpuBufferSlice allocateGpu(long size, long alignment, @GpuBuffer.Usage int usage, long minimumAllocation, long elementSize) {
+        public GpuBufferSlice allocateGpu(long size, long alignment, int usage, @GpuBuffer.Usage long minimumAllocation, long elementSize) {
             assert size <= 2147483647L;
             TransientBlockAllocator.Allocation<GlTransientMemory.PersistentMapping.GlAllocation> alloc = this.gpuBlockAllocator
                 .allocate(size, alignment, minimumAllocation, elementSize);
@@ -411,11 +_,7 @@
             return uploadedBuffers;
         }
 
-        private record GlAllocation(int glBuffer, long hostPtr, long size) implements TransientBlockAllocator.Allocator.Block {
-            @Override
-            public boolean suboptimal() {
-                return false;
-            }
+        private record GlAllocation(int glBuffer, long hostPtr, long size) {
         }
 
         private record Rotation(Runnable staging, Runnable gpu, Runnable gpuMapped) implements Runnable {
