--- a/com/mojang/renderpearl/backend/opengl/GlUtil.java
+++ b/com/mojang/renderpearl/backend/opengl/GlUtil.java
@@ -11,10 +_,8 @@
             return 34962;
         } else if ((usage & 64) != 0) {
             return 34963;
-        } else if ((usage & 128) != 0) {
-            return 35345;
         } else {
-            return (usage & 512) != 0 ? 36671 : 36663;
+            return (usage & 128) != 0 ? 35345 : 36663;
         }
     }
 }
