From a4ebcb7ed0eb29a76ad14bd0858344639b421c99 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Thu, 27 Jun 2013 17:26:09 +1000
Subject: [PATCH] Properly Close Inventories

Properly close inventories when unloading and switching worlds.

diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 3e3cb81e0..c48b60273 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1273,6 +1273,15 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     }
 
     public void unload(LevelChunk levelChunk) {
+        // Spigot Start
+        for (net.minecraft.world.level.block.entity.BlockEntity blockentity : levelChunk.getBlockEntities().values()) {
+            if (blockentity instanceof net.minecraft.world.Container) {
+                for (org.bukkit.entity.HumanEntity h : Lists.newArrayList(((net.minecraft.world.Container) blockentity).getViewers())) {
+                    h.closeInventory();
+                }
+            }
+        }
+        // Spigot End
         levelChunk.clearAllBlockEntities();
         levelChunk.unregisterTickContainerFromLevel(this);
         this.debugSynchronizers.dropChunk(levelChunk.getPos());
@@ -2248,6 +2257,13 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
         public void onTrackingEnd(Entity entity) {
             org.spigotmc.AsyncCatcher.catchOp("entity unregister"); // Spigot
+            // Spigot Start
+            if (entity.getBukkitEntity() instanceof org.bukkit.inventory.InventoryHolder && (!(entity instanceof ServerPlayer) || entity.getRemovalReason() != Entity.RemovalReason.KILLED)) { // SPIGOT-6876: closeInventory clears death message
+                for (org.bukkit.entity.HumanEntity h : Lists.newArrayList(((org.bukkit.inventory.InventoryHolder) entity.getBukkitEntity()).getInventory().getViewers())) {
+                    h.closeInventory();
+                }
+            }
+            // Spigot End
             ServerLevel.this.getChunkSource().removeEntity(entity);
             if (entity instanceof ServerPlayer serverplayer) {
                 ServerLevel.this.players.remove(serverplayer);
-- 
2.53.0

