From 1dfcf789e8c040ec11e7ada2e017f8c55b4eb9f7 Mon Sep 17 00:00:00 2001
From: md_5 <git@md-5.net>
Date: Sun, 13 Apr 2014 09:00:59 +1000
Subject: [PATCH] Print Stack on InternalException


diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index a0f9b11b4..7e3d6c4fc 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -155,6 +155,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
 
             }
         }
+        if (net.minecraft.server.MinecraftServer.getServer().isDebugging()) cause.printStackTrace(); // Spigot
     }
 
     protected void channelRead0(ChannelHandlerContext ctx, Packet<?> packet) {
-- 
2.53.0

