From 081b838eb75efda9e04099545b9e3302b6982bba Mon Sep 17 00:00:00 2001
From: ninja <xninja@openmailbox.org>
Date: Tue, 8 Apr 2014 14:05:19 +0200
Subject: [PATCH] Implement PlayerSpawnLocationEvent.


diff --git a/src/main/java/net/minecraft/server/network/config/PrepareSpawnTask.java b/src/main/java/net/minecraft/server/network/config/PrepareSpawnTask.java
index 5575ff12c..01cf12024 100644
--- a/src/main/java/net/minecraft/server/network/config/PrepareSpawnTask.java
+++ b/src/main/java/net/minecraft/server/network/config/PrepareSpawnTask.java
@@ -32,6 +32,11 @@ import org.slf4j.Logger;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.world.level.Level;
 // CraftBukkit end
+// Spigot start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.entity.Player;
+// Spigot end
 
 public class PrepareSpawnTask implements ConfigurationTask {
 
@@ -226,6 +231,18 @@ public class PrepareSpawnTask implements ConfigurationTask {
                 Objects.requireNonNull(serverplayer);
                 optional.ifPresent(serverplayer::load);
                 serverplayer.snapTo(this.spawnPosition, this.spawnAngle.x, this.spawnAngle.y);
+                // Spigot start - spawn location event
+                Player spawnPlayer = serverplayer.getBukkitEntity();
+                org.spigotmc.event.player.PlayerSpawnLocationEvent ev = new org.spigotmc.event.player.PlayerSpawnLocationEvent(spawnPlayer, spawnPlayer.getLocation());
+                spawnPlayer.getServer().getPluginManager().callEvent(ev);
+
+                Location loc = ev.getSpawnLocation();
+                ServerLevel spawnWorld = ((CraftWorld) loc.getWorld()).getHandle();
+
+                serverplayer.spawnIn(spawnWorld, true);
+                serverplayer.gameMode.setLevel((ServerLevel) serverplayer.level());
+                serverplayer.absSnapTo(loc.getX(), loc.getY(), loc.getZ(), loc.getYaw(), loc.getPitch());
+                // Spigot end
                 PrepareSpawnTask.this.server.getPlayerList().placeNewPlayer(connection, serverplayer, cookie);
                 optional.ifPresent((valueinput) -> {
                     serverplayer.loadAndSpawnEnderPearls(valueinput);
-- 
2.53.0

