--- a/net/minecraft/client/renderer/entity/state/HumanoidRenderState.java
+++ b/net/minecraft/client/renderer/entity/state/HumanoidRenderState.java
@@ -28,6 +_,6 @@
 
     @Override
     public float ticksUsingItem(HumanoidArm arm) {
-        return this.isUsingItem && this.useItemHand.asArm(this.mainArm) == arm ? this.ticksUsingItem : 0.0F;
+        return this.isUsingItem && this.useItemHand == InteractionHand.MAIN_HAND == (arm == this.mainArm) ? this.ticksUsingItem : 0.0F;
     }
 }
