--- a/net/minecraft/client/particle/AshParticle.java
+++ b/net/minecraft/client/particle/AshParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class AshParticle extends BaseAshSmokeParticle {
-    public AshParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
+    protected AshParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
         super(level, x, y, z, 0.1F, -0.1F, 0.1F, xa, ya, za, scale, sprites, 0.5F, 20, 0.1F, false);
     }
 
