--- a/net/minecraft/client/particle/AttackSweepParticle.java
+++ b/net/minecraft/client/particle/AttackSweepParticle.java
@@ -10,7 +_,7 @@
 public class AttackSweepParticle extends SingleQuadParticle {
     private final SpriteSet sprites;
 
-    public AttackSweepParticle(ClientLevel level, double x, double y, double z, double size, SpriteSet sprites) {
+    private AttackSweepParticle(ClientLevel level, double x, double y, double z, double size, SpriteSet sprites) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprites.first());
         this.sprites = sprites;
         this.lifetime = 4;
