--- a/net/minecraft/client/particle/AttackSweepParticle.java
+++ b/net/minecraft/client/particle/AttackSweepParticle.java
@@ -9,7 +_,7 @@
 public class AttackSweepParticle extends TextureSheetParticle {
     private final SpriteSet sprites;
 
-    AttackSweepParticle(ClientLevel p_105546_, double p_105547_, double p_105548_, double p_105549_, double p_105550_, SpriteSet p_105551_) {
+    protected AttackSweepParticle(ClientLevel p_105546_, double p_105547_, double p_105548_, double p_105549_, double p_105550_, SpriteSet p_105551_) {
         super(p_105546_, p_105547_, p_105548_, p_105549_, 0.0, 0.0, 0.0);
         this.sprites = p_105551_;
         this.lifetime = 4;
