--- a/net/minecraft/client/particle/BlockMarker.java
+++ b/net/minecraft/client/particle/BlockMarker.java
@@ -9,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class BlockMarker extends TextureSheetParticle {
-    BlockMarker(ClientLevel p_194267_, double p_194268_, double p_194269_, double p_194270_, BlockState p_194271_) {
+    protected BlockMarker(ClientLevel p_194267_, double p_194268_, double p_194269_, double p_194270_, BlockState p_194271_) {
         super(p_194267_, p_194268_, p_194269_, p_194270_);
         this.setSprite(Minecraft.getInstance().getBlockRenderer().getBlockModelShaper().getParticleIcon(p_194271_));
         this.gravity = 0.0F;
