--- a/net/minecraft/client/particle/BlockMarker.java
+++ b/net/minecraft/client/particle/BlockMarker.java
@@ -12,7 +_,7 @@
 public class BlockMarker extends SingleQuadParticle {
     private final SingleQuadParticle.Layer layer;
 
-    public BlockMarker(ClientLevel level, double x, double y, double z, BlockState state) {
+    private BlockMarker(ClientLevel level, double x, double y, double z, BlockState state) {
         super(level, x, y, z, Minecraft.getInstance().getModelManager().getBlockStateModelSet().getParticleMaterial(state).sprite());
         this.gravity = 0.0F;
         this.lifetime = 80;
