--- a/net/minecraft/client/particle/BreakingItemParticle.java
+++ b/net/minecraft/client/particle/BreakingItemParticle.java
@@ -22,7 +_,7 @@
     private final float vo;
     private final SingleQuadParticle.Layer layer;
 
-    public BreakingItemParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
+    private BreakingItemParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
         this(level, x, y, z, sprite);
         this.xd *= 0.1F;
         this.yd *= 0.1F;
@@ -32,7 +_,7 @@
         this.zd += za;
     }
 
-    public BreakingItemParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
+    protected BreakingItemParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprite);
         this.gravity = 1.0F;
         this.quadSize /= 2.0F;
