--- a/net/minecraft/client/particle/BubblePopParticle.java
+++ b/net/minecraft/client/particle/BubblePopParticle.java
@@ -9,7 +_,7 @@
 public class BubblePopParticle extends TextureSheetParticle {
     private final SpriteSet sprites;
 
-    BubblePopParticle(
+    protected BubblePopParticle(
         ClientLevel p_105814_, double p_105815_, double p_105816_, double p_105817_, double p_105818_, double p_105819_, double p_105820_, SpriteSet p_105821_
     ) {
         super(p_105814_, p_105815_, p_105816_, p_105817_);
