--- a/net/minecraft/client/particle/BubbleParticle.java
+++ b/net/minecraft/client/particle/BubbleParticle.java
@@ -11,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class BubbleParticle extends SingleQuadParticle {
-    public BubbleParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
+    private BubbleParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
         super(level, x, y, z, sprite);
         this.setSize(0.02F, 0.02F);
         this.quadSize = this.quadSize * (this.random.nextFloat() * 0.6F + 0.2F);
