--- a/net/minecraft/client/particle/BubbleColumnUpParticle.java
+++ b/net/minecraft/client/particle/BubbleColumnUpParticle.java
@@ -9,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class BubbleColumnUpParticle extends TextureSheetParticle {
-    BubbleColumnUpParticle(ClientLevel p_105733_, double p_105734_, double p_105735_, double p_105736_, double p_105737_, double p_105738_, double p_105739_) {
+    protected BubbleColumnUpParticle(ClientLevel p_105733_, double p_105734_, double p_105735_, double p_105736_, double p_105737_, double p_105738_, double p_105739_) {
         super(p_105733_, p_105734_, p_105735_, p_105736_);
         this.gravity = -0.125F;
         this.friction = 0.85F;
