--- a/net/minecraft/client/particle/BubbleColumnUpParticle.java
+++ b/net/minecraft/client/particle/BubbleColumnUpParticle.java
@@ -11,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class BubbleColumnUpParticle extends SingleQuadParticle {
-    public BubbleColumnUpParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
+    private BubbleColumnUpParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
         super(level, x, y, z, sprite);
         this.gravity = -0.125F;
         this.friction = 0.85F;
