--- a/net/minecraft/client/particle/CampfireSmokeParticle.java
+++ b/net/minecraft/client/particle/CampfireSmokeParticle.java
@@ -9,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class CampfireSmokeParticle extends SingleQuadParticle {
-    public CampfireSmokeParticle(
+    private CampfireSmokeParticle(
         ClientLevel level, double x, double y, double z, double xa, double ya, double za, boolean isSignalFire, TextureAtlasSprite sprite
     ) {
         super(level, x, y, z, sprite);
