--- a/net/minecraft/client/particle/CampfireSmokeParticle.java
+++ b/net/minecraft/client/particle/CampfireSmokeParticle.java
@@ -7,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class CampfireSmokeParticle extends TextureSheetParticle {
-    CampfireSmokeParticle(
+    protected CampfireSmokeParticle(
         ClientLevel p_105856_, double p_105857_, double p_105858_, double p_105859_, double p_105860_, double p_105861_, double p_105862_, boolean p_105863_
     ) {
         super(p_105856_, p_105857_, p_105858_, p_105859_);
