--- a/net/minecraft/client/particle/CritParticle.java
+++ b/net/minecraft/client/particle/CritParticle.java
@@ -10,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class CritParticle extends SingleQuadParticle {
-    public CritParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
+    private CritParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprite);
         this.friction = 0.7F;
         this.gravity = 0.5F;
