--- a/net/minecraft/client/particle/CritParticle.java
+++ b/net/minecraft/client/particle/CritParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class CritParticle extends TextureSheetParticle {
-    CritParticle(ClientLevel p_105919_, double p_105920_, double p_105921_, double p_105922_, double p_105923_, double p_105924_, double p_105925_) {
+    protected CritParticle(ClientLevel p_105919_, double p_105920_, double p_105921_, double p_105922_, double p_105923_, double p_105924_, double p_105925_) {
         super(p_105919_, p_105920_, p_105921_, p_105922_, 0.0, 0.0, 0.0);
         this.friction = 0.7F;
         this.gravity = 0.5F;
