--- a/net/minecraft/client/particle/DragonBreathParticle.java
+++ b/net/minecraft/client/particle/DragonBreathParticle.java
@@ -20,7 +_,7 @@
     private boolean hasHitGround;
     private final SpriteSet sprites;
 
-    public DragonBreathParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
+    private DragonBreathParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
         super(level, x, y, z, sprites.first());
         this.friction = 0.96F;
         this.xd = xa;
