--- a/net/minecraft/client/particle/DragonBreathParticle.java
+++ b/net/minecraft/client/particle/DragonBreathParticle.java
@@ -19,7 +_,7 @@
     private boolean hasHitGround;
     private final SpriteSet sprites;
 
-    DragonBreathParticle(
+    protected DragonBreathParticle(
         ClientLevel p_106005_, double p_106006_, double p_106007_, double p_106008_, double p_106009_, double p_106010_, double p_106011_, SpriteSet p_106012_
     ) {
         super(p_106005_, p_106006_, p_106007_, p_106008_);
