--- a/net/minecraft/client/particle/DripParticle.java
+++ b/net/minecraft/client/particle/DripParticle.java
@@ -23,7 +_,7 @@
     private final Fluid type;
     protected boolean isGlowing;
 
-    public DripParticle(ClientLevel level, double x, double y, double z, Fluid type, TextureAtlasSprite sprite) {
+    private DripParticle(ClientLevel level, double x, double y, double z, Fluid type, TextureAtlasSprite sprite) {
         super(level, x, y, z, sprite);
         this.setSize(0.01F, 0.01F);
         this.gravity = 0.06F;
@@ -79,7 +_,7 @@
     }
 
     private static class CoolingDripHangParticle extends DripParticle.DripHangParticle {
-        public CoolingDripHangParticle(ClientLevel level, double x, double y, double z, Fluid type, ParticleOptions fallingParticle, TextureAtlasSprite sprite) {
+        private CoolingDripHangParticle(ClientLevel level, double x, double y, double z, Fluid type, ParticleOptions fallingParticle, TextureAtlasSprite sprite) {
             super(level, x, y, z, type, fallingParticle, sprite);
         }
 
@@ -95,7 +_,7 @@
     private static class DripHangParticle extends DripParticle {
         private final ParticleOptions fallingParticle;
 
-        public DripHangParticle(ClientLevel level, double x, double y, double z, Fluid type, ParticleOptions fallingParticle, TextureAtlasSprite sprite) {
+        private DripHangParticle(ClientLevel level, double x, double y, double z, Fluid type, ParticleOptions fallingParticle, TextureAtlasSprite sprite) {
             super(level, x, y, z, type, sprite);
             this.fallingParticle = fallingParticle;
             this.gravity *= 0.02F;
@@ -119,14 +_,14 @@
     }
 
     private static class DripLandParticle extends DripParticle {
-        public DripLandParticle(ClientLevel level, double x, double y, double z, Fluid type, TextureAtlasSprite sprite) {
+        private DripLandParticle(ClientLevel level, double x, double y, double z, Fluid type, TextureAtlasSprite sprite) {
             super(level, x, y, z, type, sprite);
             this.lifetime = (int)(16.0 / (this.random.nextFloat() * 0.8 + 0.2));
         }
     }
 
     private static class DripstoneFallAndLandParticle extends DripParticle.FallAndLandParticle {
-        public DripstoneFallAndLandParticle(
+        private DripstoneFallAndLandParticle(
             ClientLevel level, double x, double y, double z, Fluid type, ParticleOptions landParticle, TextureAtlasSprite sprite
         ) {
             super(level, x, y, z, type, landParticle, sprite);
@@ -213,7 +_,7 @@
     private static class FallAndLandParticle extends DripParticle.FallingParticle {
         protected final ParticleOptions landParticle;
 
-        public FallAndLandParticle(ClientLevel level, double x, double y, double z, Fluid type, ParticleOptions landParticle, TextureAtlasSprite sprite) {
+        private FallAndLandParticle(ClientLevel level, double x, double y, double z, Fluid type, ParticleOptions landParticle, TextureAtlasSprite sprite) {
             super(level, x, y, z, type, sprite);
             this.lifetime = (int)(64.0 / (this.random.nextFloat() * 0.8 + 0.2));
             this.landParticle = landParticle;
@@ -229,7 +_,7 @@
     }
 
     private static class FallingParticle extends DripParticle {
-        public FallingParticle(ClientLevel level, double x, double y, double z, Fluid type, TextureAtlasSprite sprite) {
+        private FallingParticle(ClientLevel level, double x, double y, double z, Fluid type, TextureAtlasSprite sprite) {
             super(level, x, y, z, type, sprite);
         }
 
@@ -242,7 +_,7 @@
     }
 
     private static class HoneyFallAndLandParticle extends DripParticle.FallAndLandParticle {
-        public HoneyFallAndLandParticle(ClientLevel level, double x, double y, double z, Fluid type, ParticleOptions landParticle, TextureAtlasSprite sprite) {
+        private HoneyFallAndLandParticle(ClientLevel level, double x, double y, double z, Fluid type, ParticleOptions landParticle, TextureAtlasSprite sprite) {
             super(level, x, y, z, type, landParticle, sprite);
         }
 
