--- a/net/minecraft/client/particle/DripParticle.java
+++ b/net/minecraft/client/particle/DripParticle.java
@@ -20,7 +_,7 @@
     private final Fluid type;
     protected boolean isGlowing;
 
-    DripParticle(ClientLevel p_106051_, double p_106052_, double p_106053_, double p_106054_, Fluid p_106055_) {
+    protected DripParticle(ClientLevel p_106051_, double p_106052_, double p_106053_, double p_106054_, Fluid p_106055_) {
         super(p_106051_, p_106052_, p_106053_, p_106054_);
         this.setSize(0.01F, 0.01F);
         this.gravity = 0.06F;
