--- a/net/minecraft/client/particle/DustParticleBase.java
+++ b/net/minecraft/client/particle/DustParticleBase.java
@@ -10,7 +_,7 @@
 public class DustParticleBase<T extends ScalableParticleOptionsBase> extends SingleQuadParticle {
     private final SpriteSet sprites;
 
-    public DustParticleBase(ClientLevel level, double x, double y, double z, double xAux, double yAux, double zAux, T options, SpriteSet sprites) {
+    protected DustParticleBase(ClientLevel level, double x, double y, double z, double xAux, double yAux, double zAux, T options, SpriteSet sprites) {
         super(level, x, y, z, xAux, yAux, zAux, sprites.first());
         this.friction = 0.96F;
         this.speedUpWhenYMotionIsBlocked = true;
