--- a/net/minecraft/client/particle/DustParticle.java
+++ b/net/minecraft/client/particle/DustParticle.java
@@ -9,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class DustParticle extends DustParticleBase<DustParticleOptions> {
-    public DustParticle(
+    protected DustParticle(
         ClientLevel level, double x, double y, double z, double xAux, double yAux, double zAux, DustParticleOptions options, SpriteSet sprites
     ) {
         super(level, x, y, z, xAux, yAux, zAux, options, sprites);
