--- a/net/minecraft/client/particle/DustColorTransitionParticle.java
+++ b/net/minecraft/client/particle/DustColorTransitionParticle.java
@@ -14,7 +_,7 @@
     private final Vector3f fromColor;
     private final Vector3f toColor;
 
-    public DustColorTransitionParticle(
+    protected DustColorTransitionParticle(
         ClientLevel level, double x, double y, double z, double xAux, double yAux, double zAux, DustColorTransitionOptions options, SpriteSet sprites
     ) {
         super(level, x, y, z, xAux, yAux, zAux, options, sprites);
