--- a/net/minecraft/client/particle/DustPlumeParticle.java
+++ b/net/minecraft/client/particle/DustPlumeParticle.java
@@ -11,7 +_,7 @@
 public class DustPlumeParticle extends BaseAshSmokeParticle {
     private static final int COLOR_RGB24 = 12235202;
 
-    public DustPlumeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
+    protected DustPlumeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
         super(level, x, y, z, 0.7F, 0.6F, 0.7F, xa, ya + 0.15F, za, scale, sprites, 0.5F, 7, 0.5F, false);
         float colorShift = this.random.nextFloat() * 0.2F;
         this.rCol = ARGB.red(12235202) / 255.0F - colorShift;
