--- a/net/minecraft/client/particle/ElderGuardianParticle.java
+++ b/net/minecraft/client/particle/ElderGuardianParticle.java
@@ -17,7 +_,7 @@
     protected final GuardianParticleModel model;
     protected final RenderType renderType = RenderTypes.entityTranslucent(ElderGuardianRenderer.GUARDIAN_ELDER_LOCATION);
 
-    public ElderGuardianParticle(ClientLevel level, double x, double y, double z) {
+    private ElderGuardianParticle(ClientLevel level, double x, double y, double z) {
         super(level, x, y, z);
         this.model = new GuardianParticleModel(Minecraft.getInstance().getEntityModels().bakeLayer(ModelLayers.ELDER_GUARDIAN));
         this.gravity = 0.0F;
