--- a/net/minecraft/client/particle/ElderGuardianParticleGroup.java
+++ b/net/minecraft/client/particle/ElderGuardianParticleGroup.java
@@ -42,8 +_,8 @@
             int color = ARGB.colorFromFloat(alpha, 1.0F, 1.0F, 1.0F);
             PoseStack poseStack = new PoseStack();
             poseStack.pushPose();
-            poseStack.rotate(camera.rotation());
-            poseStack.rotateDegrees(Axis.XP, 60.0F - 150.0F * ageScale);
+            poseStack.mulPose(camera.rotation());
+            poseStack.mulPose(Axis.XP.rotationDegrees(60.0F - 150.0F * ageScale));
             float scale = 0.42553192F;
             poseStack.scale(0.42553192F, -0.42553192F, -0.42553192F);
             poseStack.translate(0.0F, -0.56F, 3.5F);
