--- a/net/minecraft/client/particle/EndRodParticle.java
+++ b/net/minecraft/client/particle/EndRodParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class EndRodParticle extends SimpleAnimatedParticle {
-    public EndRodParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
+    private EndRodParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
         super(level, x, y, z, sprites, 0.0125F);
         this.xd = xa;
         this.yd = ya;
