--- a/net/minecraft/client/particle/EndRodParticle.java
+++ b/net/minecraft/client/particle/EndRodParticle.java
@@ -7,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class EndRodParticle extends SimpleAnimatedParticle {
-    EndRodParticle(
+    protected EndRodParticle(
         ClientLevel p_106531_, double p_106532_, double p_106533_, double p_106534_, double p_106535_, double p_106536_, double p_106537_, SpriteSet p_106538_
     ) {
         super(p_106531_, p_106532_, p_106533_, p_106534_, p_106538_, 0.0125F);
