--- a/net/minecraft/client/particle/ExplodeParticle.java
+++ b/net/minecraft/client/particle/ExplodeParticle.java
@@ -10,7 +_,7 @@
 public class ExplodeParticle extends SingleQuadParticle {
     private final SpriteSet sprites;
 
-    public ExplodeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
+    protected ExplodeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
         super(level, x, y, z, sprites.first());
         this.gravity = -0.1F;
         this.friction = 0.9F;
