--- a/net/minecraft/client/particle/FallingDustParticle.java
+++ b/net/minecraft/client/particle/FallingDustParticle.java
@@ -19,7 +_,7 @@
     private final float rotSpeed;
     private final SpriteSet sprites;
 
-    public FallingDustParticle(ClientLevel level, double x, double y, double z, float r, float g, float b, SpriteSet sprites) {
+    private FallingDustParticle(ClientLevel level, double x, double y, double z, float r, float g, float b, SpriteSet sprites) {
         super(level, x, y, z, sprites.first());
         this.sprites = sprites;
         this.rCol = r;
