--- a/net/minecraft/client/particle/FallingDustParticle.java
+++ b/net/minecraft/client/particle/FallingDustParticle.java
@@ -17,7 +_,7 @@
     private final float rotSpeed;
     private final SpriteSet sprites;
 
-    FallingDustParticle(
+    protected FallingDustParticle(
         ClientLevel p_106610_, double p_106611_, double p_106612_, double p_106613_, float p_106614_, float p_106615_, float p_106616_, SpriteSet p_106617_
     ) {
         super(p_106610_, p_106611_, p_106612_, p_106613_);
