--- a/net/minecraft/client/particle/FireflyParticle.java
+++ b/net/minecraft/client/particle/FireflyParticle.java
@@ -18,7 +_,7 @@
     private static final int PARTICLE_MIN_LIFETIME = 200;
     private static final int PARTICLE_MAX_LIFETIME = 300;
 
-    public FireflyParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
+    private FireflyParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
         super(level, x, y, z, xa, ya, za, sprite);
         this.speedUpWhenYMotionIsBlocked = true;
         this.friction = 0.96F;
