--- a/net/minecraft/client/particle/FireworkParticles.java
+++ b/net/minecraft/client/particle/FireworkParticles.java
@@ -49,7 +_,7 @@
 
     @OnlyIn(Dist.CLIENT)
     public static class OverlayParticle extends TextureSheetParticle {
-        OverlayParticle(ClientLevel p_106677_, double p_106678_, double p_106679_, double p_106680_) {
+        protected OverlayParticle(ClientLevel p_106677_, double p_106678_, double p_106679_, double p_106680_) {
             super(p_106677_, p_106678_, p_106679_, p_106680_);
             this.lifetime = 4;
         }
@@ -81,7 +_,7 @@
         private float fadeB;
         private boolean hasFade;
 
-        SparkParticle(
+        protected SparkParticle(
             ClientLevel p_106702_,
             double p_106703_,
             double p_106704_,
@@ -253,6 +_,10 @@
                     intlist = IntList.of(DyeColor.BLACK.getFireworkColor());
                 }
 
+                var factory = net.neoforged.neoforge.client.FireworkShapeFactoryRegistry.get(fireworkexplosion1.shape());
+                if (factory != null)
+                    factory.build(this, flag3, flag4, intlist.toIntArray(), intlist1.toIntArray());
+                else
                 switch (fireworkexplosion1.shape()) {
                     case SMALL_BALL:
                         this.createParticleBall(0.25, 2, intlist, intlist1, flag3, flag4);
