--- a/net/minecraft/client/particle/FireworkParticles.java
+++ b/net/minecraft/client/particle/FireworkParticles.java
@@ -41,7 +_,7 @@
     }
 
     public static class OverlayParticle extends SingleQuadParticle {
-        public OverlayParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
+        private OverlayParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
             super(level, x, y, z, sprite);
             this.lifetime = 4;
         }
@@ -72,7 +_,7 @@
         private float fadeB;
         private boolean hasFade;
 
-        public SparkParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, ParticleEngine engine, SpriteSet sprites) {
+        private SparkParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, ParticleEngine engine, SpriteSet sprites) {
             super(level, x, y, z, sprites, 0.1F);
             this.xd = xa;
             this.yd = ya;
@@ -255,7 +_,7 @@
             return instance.gameRenderer.mainCamera().position().distanceToSqr(this.x, this.y, this.z) >= 256.0;
         }
 
-        public void createParticle(
+        private void createParticle(
             double x, double y, double z, double xa, double ya, double za, IntList rgbColors, IntList fadeColors, boolean trail, boolean twinkle
         ) {
             FireworkParticles.SparkParticle sparkParticle = (FireworkParticles.SparkParticle)this.engine
@@ -269,7 +_,7 @@
             }
         }
 
-        public void createParticleBall(double baseSpeed, int steps, IntList rgbColors, IntList fadeColors, boolean trail, boolean twinkle) {
+        private void createParticleBall(double baseSpeed, int steps, IntList rgbColors, IntList fadeColors, boolean trail, boolean twinkle) {
             double xx = this.x;
             double yy = this.y;
             double zz = this.z;
@@ -290,7 +_,7 @@
             }
         }
 
-        public void createParticleShape(
+        private void createParticleShape(
             double baseSpeed, double[][] coords, IntList rgbColors, IntList fadeColors, boolean trail, boolean twinkle, boolean flat
         ) {
             double sx = coords[0][0];
@@ -325,7 +_,7 @@
             }
         }
 
-        public void createParticleBurst(IntList rgbColors, IntList fadeColors, boolean trail, boolean twinkle) {
+        private void createParticleBurst(IntList rgbColors, IntList fadeColors, boolean trail, boolean twinkle) {
             double baseOffX = this.random.nextGaussian() * 0.05;
             double baseOffZ = this.random.nextGaussian() * 0.05;
 
