--- a/net/minecraft/client/particle/FlameParticle.java
+++ b/net/minecraft/client/particle/FlameParticle.java
@@ -10,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class FlameParticle extends RisingParticle {
-    public FlameParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
+    private FlameParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
         super(level, x, y, z, xd, yd, zd, sprite);
     }
 
