--- a/net/minecraft/client/particle/FlyStraightTowardsParticle.java
+++ b/net/minecraft/client/particle/FlyStraightTowardsParticle.java
@@ -18,7 +_,7 @@
     private final int startColor;
     private final int endColor;
 
-    public FlyStraightTowardsParticle(
+    private FlyStraightTowardsParticle(
         ClientLevel level, double x, double y, double z, double xd, double yd, double zd, int startColor, int endColor, TextureAtlasSprite sprite
     ) {
         super(level, x, y, z, sprite);
