--- a/net/minecraft/client/particle/FlyTowardsPositionParticle.java
+++ b/net/minecraft/client/particle/FlyTowardsPositionParticle.java
@@ -18,11 +_,11 @@
     private final boolean isGlowing;
     private final Particle.LifetimeAlpha lifetimeAlpha;
 
-    public FlyTowardsPositionParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
+    private FlyTowardsPositionParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
         this(level, x, y, z, xd, yd, zd, false, Particle.LifetimeAlpha.ALWAYS_OPAQUE, sprite);
     }
 
-    public FlyTowardsPositionParticle(
+    private FlyTowardsPositionParticle(
         ClientLevel level,
         double x,
         double y,
