--- a/net/minecraft/client/particle/FlyTowardsPositionParticle.java
+++ b/net/minecraft/client/particle/FlyTowardsPositionParticle.java
@@ -15,7 +_,7 @@
     private final boolean isGlowing;
     private final Particle.LifetimeAlpha lifetimeAlpha;
 
-    FlyTowardsPositionParticle(
+    protected FlyTowardsPositionParticle(
         ClientLevel p_323658_, double p_324523_, double p_324079_, double p_324377_, double p_324144_, double p_324286_, double p_323511_
     ) {
         this(p_323658_, p_324523_, p_324079_, p_324377_, p_324144_, p_324286_, p_323511_, false, Particle.LifetimeAlpha.ALWAYS_OPAQUE);
