--- a/net/minecraft/client/particle/GeyserBaseParticle.java
+++ b/net/minecraft/client/particle/GeyserBaseParticle.java
@@ -16,7 +_,7 @@
     private static final int MAX_LIFETIME = 25;
     private static final float FRICTION = 0.725F;
 
-    public GeyserBaseParticle(
+    private GeyserBaseParticle(
         ClientLevel level, double x, double y, double z, double xAux, double yAux, double zAux, int waterBlocks, float burstImpulseBase, SpriteSet sprites
     ) {
         float burstImpulse = burstImpulseBase + 0.25F * waterBlocks;
