--- a/net/minecraft/client/particle/GeyserEruptionParticle.java
+++ b/net/minecraft/client/particle/GeyserEruptionParticle.java
@@ -25,7 +_,7 @@
     private final GeyserBaseParticleOptions baseParticle;
     private final GeyserBaseParticleOptions poofParticle;
 
-    public GeyserEruptionParticle(ClientLevel level, double x, double y, double z, double xAux, double yAux, double zAux, GeyserParticleOptions options) {
+    protected GeyserEruptionParticle(ClientLevel level, double x, double y, double z, double xAux, double yAux, double zAux, GeyserParticleOptions options) {
         super(level, x, y, z);
         this.xa = xAux;
         this.ya = yAux;
