--- a/net/minecraft/client/particle/GeyserPlumeParticle.java
+++ b/net/minecraft/client/particle/GeyserPlumeParticle.java
@@ -29,7 +_,7 @@
     private final float maxSize;
     private boolean done;
 
-    public GeyserPlumeParticle(
+    private GeyserPlumeParticle(
         ClientLevel level, double x, double y, double z, double xa, double ya, double za, GeyserParticleOptions options, SpriteSet sprites
     ) {
         super(level, x, y, z, xa, ya, za, sprites.first());
