--- a/net/minecraft/client/particle/GlowParticle.java
+++ b/net/minecraft/client/particle/GlowParticle.java
@@ -12,7 +_,7 @@
     static final RandomSource RANDOM = RandomSource.create();
     private final SpriteSet sprites;
 
-    GlowParticle(
+    protected GlowParticle(
         ClientLevel p_172136_, double p_172137_, double p_172138_, double p_172139_, double p_172140_, double p_172141_, double p_172142_, SpriteSet p_172143_
     ) {
         super(p_172136_, p_172137_, p_172138_, p_172139_, p_172140_, p_172141_, p_172142_);
