--- a/net/minecraft/client/particle/GlowParticle.java
+++ b/net/minecraft/client/particle/GlowParticle.java
@@ -11,7 +_,7 @@
 public class GlowParticle extends SingleQuadParticle {
     private final SpriteSet sprites;
 
-    public GlowParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
+    private GlowParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
         super(level, x, y, z, xa, ya, za, sprites.first());
         this.friction = 0.96F;
         this.speedUpWhenYMotionIsBlocked = true;
