--- a/net/minecraft/client/particle/GustSeedParticle.java
+++ b/net/minecraft/client/particle/GustSeedParticle.java
@@ -12,7 +_,7 @@
     private final double scale;
     private final int tickDelayInBetween;
 
-    public GustSeedParticle(ClientLevel level, double x, double y, double z, double scale, int lifetime, int tickDelayInBetween) {
+    private GustSeedParticle(ClientLevel level, double x, double y, double z, double scale, int lifetime, int tickDelayInBetween) {
         super(level, x, y, z, 0.0, 0.0, 0.0);
         this.scale = scale;
         this.lifetime = lifetime;
