--- a/net/minecraft/client/particle/GustSeedParticle.java
+++ b/net/minecraft/client/particle/GustSeedParticle.java
@@ -11,7 +_,7 @@
     private final double scale;
     private final int tickDelayInBetween;
 
-    GustSeedParticle(ClientLevel p_311842_, double p_312103_, double p_312358_, double p_312614_, double p_326153_, int p_326148_, int p_326155_) {
+    protected GustSeedParticle(ClientLevel p_311842_, double p_312103_, double p_312358_, double p_312614_, double p_326153_, int p_326148_, int p_326155_) {
         super(p_311842_, p_312103_, p_312358_, p_312614_, 0.0, 0.0, 0.0);
         this.scale = p_326153_;
         this.lifetime = p_326148_;
