--- a/net/minecraft/client/particle/GustParticle.java
+++ b/net/minecraft/client/particle/GustParticle.java
@@ -10,7 +_,7 @@
 public class GustParticle extends SingleQuadParticle {
     private final SpriteSet sprites;
 
-    public GustParticle(ClientLevel level, double x, double y, double z, SpriteSet sprites) {
+    protected GustParticle(ClientLevel level, double x, double y, double z, SpriteSet sprites) {
         super(level, x, y, z, sprites.first());
         this.sprites = sprites;
         this.setSpriteFromAge(sprites);
