--- a/net/minecraft/client/particle/HeartParticle.java
+++ b/net/minecraft/client/particle/HeartParticle.java
@@ -10,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class HeartParticle extends SingleQuadParticle {
-    public HeartParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
+    private HeartParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprite);
         this.speedUpWhenYMotionIsBlocked = true;
         this.friction = 0.86F;
