--- a/net/minecraft/client/particle/HeartParticle.java
+++ b/net/minecraft/client/particle/HeartParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class HeartParticle extends TextureSheetParticle {
-    HeartParticle(ClientLevel p_106847_, double p_106848_, double p_106849_, double p_106850_) {
+    protected HeartParticle(ClientLevel p_106847_, double p_106848_, double p_106849_, double p_106850_) {
         super(p_106847_, p_106848_, p_106849_, p_106850_, 0.0, 0.0, 0.0);
         this.speedUpWhenYMotionIsBlocked = true;
         this.friction = 0.86F;
