--- a/net/minecraft/client/particle/HugeExplosionSeedParticle.java
+++ b/net/minecraft/client/particle/HugeExplosionSeedParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class HugeExplosionSeedParticle extends NoRenderParticle {
-    HugeExplosionSeedParticle(ClientLevel p_106947_, double p_106948_, double p_106949_, double p_106950_) {
+    protected HugeExplosionSeedParticle(ClientLevel p_106947_, double p_106948_, double p_106949_, double p_106950_) {
         super(p_106947_, p_106948_, p_106949_, p_106950_, 0.0, 0.0, 0.0);
         this.lifetime = 8;
     }
