--- a/net/minecraft/client/particle/HugeExplosionSeedParticle.java
+++ b/net/minecraft/client/particle/HugeExplosionSeedParticle.java
@@ -9,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class HugeExplosionSeedParticle extends NoRenderParticle {
-    public HugeExplosionSeedParticle(ClientLevel level, double x, double y, double z) {
+    private HugeExplosionSeedParticle(ClientLevel level, double x, double y, double z) {
         super(level, x, y, z, 0.0, 0.0, 0.0);
         this.lifetime = 8;
     }
