--- a/net/minecraft/client/particle/HugeExplosionParticle.java
+++ b/net/minecraft/client/particle/HugeExplosionParticle.java
@@ -10,7 +_,7 @@
 public class HugeExplosionParticle extends SingleQuadParticle {
     private final SpriteSet sprites;
 
-    public HugeExplosionParticle(ClientLevel level, double x, double y, double z, double size, SpriteSet sprites) {
+    protected HugeExplosionParticle(ClientLevel level, double x, double y, double z, double size, SpriteSet sprites) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprites.first());
         this.lifetime = 6 + this.random.nextInt(4);
         float col = this.random.nextFloat() * 0.6F + 0.4F;
