--- a/net/minecraft/client/particle/LargeSmokeParticle.java
+++ b/net/minecraft/client/particle/LargeSmokeParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class LargeSmokeParticle extends SmokeParticle {
-    public LargeSmokeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
+    protected LargeSmokeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
         super(level, x, y, z, xa, ya, za, 2.5F, sprites);
     }
 
