--- a/net/minecraft/client/particle/LavaParticle.java
+++ b/net/minecraft/client/particle/LavaParticle.java
@@ -11,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class LavaParticle extends SingleQuadParticle {
-    public LavaParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
+    private LavaParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprite);
         this.gravity = 0.75F;
         this.friction = 0.999F;
