--- a/net/minecraft/client/particle/LavaParticle.java
+++ b/net/minecraft/client/particle/LavaParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class LavaParticle extends TextureSheetParticle {
-    LavaParticle(ClientLevel p_107074_, double p_107075_, double p_107076_, double p_107077_) {
+    protected LavaParticle(ClientLevel p_107074_, double p_107075_, double p_107076_, double p_107077_) {
         super(p_107074_, p_107075_, p_107076_, p_107077_, 0.0, 0.0, 0.0);
         this.gravity = 0.75F;
         this.friction = 0.999F;
