--- a/net/minecraft/client/particle/MobAppearanceParticle.java
+++ b/net/minecraft/client/particle/MobAppearanceParticle.java
@@ -24,7 +_,7 @@
     private final Model model;
     private final RenderType renderType = RenderType.entityTranslucent(ElderGuardianRenderer.GUARDIAN_ELDER_LOCATION);
 
-    MobAppearanceParticle(ClientLevel p_107114_, double p_107115_, double p_107116_, double p_107117_) {
+    protected MobAppearanceParticle(ClientLevel p_107114_, double p_107115_, double p_107116_, double p_107117_) {
         super(p_107114_, p_107115_, p_107116_, p_107117_);
         this.model = new GuardianModel(Minecraft.getInstance().getEntityModels().bakeLayer(ModelLayers.ELDER_GUARDIAN));
         this.gravity = 0.0F;
