--- a/net/minecraft/client/particle/NoRenderParticle.java
+++ b/net/minecraft/client/particle/NoRenderParticle.java
@@ -6,11 +_,11 @@
 
 @OnlyIn(Dist.CLIENT)
 public class NoRenderParticle extends Particle {
-    public NoRenderParticle(ClientLevel level, double x, double y, double z) {
+    protected NoRenderParticle(ClientLevel level, double x, double y, double z) {
         super(level, x, y, z);
     }
 
-    public NoRenderParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za) {
+    protected NoRenderParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za) {
         super(level, x, y, z, xa, ya, za);
     }
 
