--- a/net/minecraft/client/particle/NoteParticle.java
+++ b/net/minecraft/client/particle/NoteParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class NoteParticle extends TextureSheetParticle {
-    NoteParticle(ClientLevel p_107167_, double p_107168_, double p_107169_, double p_107170_, double p_107171_) {
+    protected NoteParticle(ClientLevel p_107167_, double p_107168_, double p_107169_, double p_107170_, double p_107171_) {
         super(p_107167_, p_107168_, p_107169_, p_107170_, 0.0, 0.0, 0.0);
         this.friction = 0.66F;
         this.speedUpWhenYMotionIsBlocked = true;
