--- a/net/minecraft/client/particle/NoteParticle.java
+++ b/net/minecraft/client/particle/NoteParticle.java
@@ -10,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class NoteParticle extends SingleQuadParticle {
-    public NoteParticle(ClientLevel level, double x, double y, double z, double color, TextureAtlasSprite sprite) {
+    private NoteParticle(ClientLevel level, double x, double y, double z, double color, TextureAtlasSprite sprite) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprite);
         this.friction = 0.66F;
         this.speedUpWhenYMotionIsBlocked = true;
