--- a/net/minecraft/client/particle/NoxiousGasCloudParticle.java
+++ b/net/minecraft/client/particle/NoxiousGasCloudParticle.java
@@ -16,7 +_,7 @@
 public class NoxiousGasCloudParticle extends NoRenderParticle {
     private static final int PARTICLE_TICKS = 2;
 
-    public NoxiousGasCloudParticle(ClientLevel level, double x, double y, double z) {
+    protected NoxiousGasCloudParticle(ClientLevel level, double x, double y, double z) {
         super(level, x, y, z);
         this.lifetime = 20;
     }
