--- a/net/minecraft/client/particle/Particle.java
+++ b/net/minecraft/client/particle/Particle.java
@@ -43,7 +_,7 @@
     protected float friction = 0.98F;
     protected boolean speedUpWhenYMotionIsBlocked = false;
 
-    public Particle(ClientLevel level, double x, double y, double z) {
+    protected Particle(ClientLevel level, double x, double y, double z) {
         this.level = level;
         this.setSize(0.2F, 0.2F);
         this.setPos(x, y, z);
