--- a/net/minecraft/client/particle/ParticleResources.java
+++ b/net/minecraft/client/particle/ParticleResources.java
@@ -186,11 +_,11 @@
         this.register(ParticleTypes.SULFUR_CUBE_GOO, BreakingItemParticle.SulfurCubeProvider::new);
     }
 
-    public <T extends ParticleOptions> void register(ParticleType<T> type, ParticleProvider<T> provider) {
+    private <T extends ParticleOptions> void register(ParticleType<T> type, ParticleProvider<T> provider) {
         this.providers.put(BuiltInRegistries.PARTICLE_TYPE.getId(type), provider);
     }
 
-    public <T extends ParticleOptions> void register(ParticleType<T> type, ParticleResources.SpriteParticleRegistration<T> provider) {
+    private <T extends ParticleOptions> void register(ParticleType<T> type, ParticleResources.SpriteParticleRegistration<T> provider) {
         ParticleResources.MutableSpriteSet spriteSet = new ParticleResources.MutableSpriteSet();
         this.spriteSets.put(BuiltInRegistries.PARTICLE_TYPE.getKey(type), spriteSet);
         this.providers.put(BuiltInRegistries.PARTICLE_TYPE.getId(type), provider.create(spriteSet));
@@ -311,7 +_,7 @@
     }
 
     @FunctionalInterface
-    public interface SpriteParticleRegistration<T extends ParticleOptions> {
+    private interface SpriteParticleRegistration<T extends ParticleOptions> {
         ParticleProvider<T> create(SpriteSet spriteSet);
     }
 }
