--- a/net/minecraft/client/particle/PlayerCloudParticle.java
+++ b/net/minecraft/client/particle/PlayerCloudParticle.java
@@ -11,7 +_,7 @@
 public class PlayerCloudParticle extends TextureSheetParticle {
     private final SpriteSet sprites;
 
-    PlayerCloudParticle(
+    protected PlayerCloudParticle(
         ClientLevel p_107483_, double p_107484_, double p_107485_, double p_107486_, double p_107487_, double p_107488_, double p_107489_, SpriteSet p_107490_
     ) {
         super(p_107483_, p_107484_, p_107485_, p_107486_, 0.0, 0.0, 0.0);
