--- a/net/minecraft/client/particle/PlayerCloudParticle.java
+++ b/net/minecraft/client/particle/PlayerCloudParticle.java
@@ -12,7 +_,7 @@
 public class PlayerCloudParticle extends SingleQuadParticle {
     private final SpriteSet sprites;
 
-    public PlayerCloudParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
+    private PlayerCloudParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprites.first());
         this.friction = 0.96F;
         this.sprites = sprites;
