--- a/net/minecraft/client/particle/ReversePortalParticle.java
+++ b/net/minecraft/client/particle/ReversePortalParticle.java
@@ -7,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class ReversePortalParticle extends PortalParticle {
-    ReversePortalParticle(ClientLevel p_107590_, double p_107591_, double p_107592_, double p_107593_, double p_107594_, double p_107595_, double p_107596_) {
+    protected ReversePortalParticle(ClientLevel p_107590_, double p_107591_, double p_107592_, double p_107593_, double p_107594_, double p_107595_, double p_107596_) {
         super(p_107590_, p_107591_, p_107592_, p_107593_, p_107594_, p_107595_, p_107596_);
         this.quadSize *= 1.5F;
         this.lifetime = (int)(Math.random() * 2.0) + 60;
@@ -31,6 +_,7 @@
             this.x = this.x + this.xd * (double)f;
             this.y = this.y + this.yd * (double)f;
             this.z = this.z + this.zd * (double)f;
+            this.setPos(this.x, this.y, this.z); // Neo: update the particle's bounding box
         }
     }
 
