--- a/net/minecraft/client/particle/ReversePortalParticle.java
+++ b/net/minecraft/client/particle/ReversePortalParticle.java
@@ -9,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class ReversePortalParticle extends PortalParticle {
-    public ReversePortalParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
+    private ReversePortalParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
         super(level, x, y, z, xd, yd, zd, sprite);
         this.quadSize *= 1.5F;
         this.lifetime = (int)(this.random.nextFloat() * 2.0F) + 60;
