--- a/net/minecraft/client/particle/RisingParticle.java
+++ b/net/minecraft/client/particle/RisingParticle.java
@@ -7,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public abstract class RisingParticle extends SingleQuadParticle {
-    public RisingParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
+    protected RisingParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
         super(level, x, y, z, xd, yd, zd, sprite);
         this.friction = 0.96F;
         this.xd = this.xd * 0.01F + xd;
