--- a/net/minecraft/client/particle/SculkChargeParticle.java
+++ b/net/minecraft/client/particle/SculkChargeParticle.java
@@ -9,7 +_,7 @@
 public class SculkChargeParticle extends TextureSheetParticle {
     private final SpriteSet sprites;
 
-    SculkChargeParticle(
+    protected SculkChargeParticle(
         ClientLevel p_233892_, double p_233893_, double p_233894_, double p_233895_, double p_233896_, double p_233897_, double p_233898_, SpriteSet p_233899_
     ) {
         super(p_233892_, p_233893_, p_233894_, p_233895_, p_233896_, p_233897_, p_233898_);
