--- a/net/minecraft/client/particle/SculkChargePopParticle.java
+++ b/net/minecraft/client/particle/SculkChargePopParticle.java
@@ -11,7 +_,7 @@
 public class SculkChargePopParticle extends SingleQuadParticle {
     private final SpriteSet sprites;
 
-    public SculkChargePopParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, SpriteSet sprites) {
+    private SculkChargePopParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, SpriteSet sprites) {
         super(level, x, y, z, xd, yd, zd, sprites.first());
         this.friction = 0.96F;
         this.sprites = sprites;
