--- a/net/minecraft/client/particle/SculkChargePopParticle.java
+++ b/net/minecraft/client/particle/SculkChargePopParticle.java
@@ -9,7 +_,7 @@
 public class SculkChargePopParticle extends TextureSheetParticle {
     private final SpriteSet sprites;
 
-    SculkChargePopParticle(
+    protected SculkChargePopParticle(
         ClientLevel p_233932_, double p_233933_, double p_233934_, double p_233935_, double p_233936_, double p_233937_, double p_233938_, SpriteSet p_233939_
     ) {
         super(p_233932_, p_233933_, p_233934_, p_233935_, p_233936_, p_233937_, p_233938_);
