--- a/net/minecraft/client/particle/ShriekParticle.java
+++ b/net/minecraft/client/particle/ShriekParticle.java
@@ -17,7 +_,7 @@
     private static final float MAGICAL_X_ROT = 1.0472F;
     private int delay;
 
-    public ShriekParticle(ClientLevel level, double x, double y, double z, int delay, TextureAtlasSprite sprite) {
+    private ShriekParticle(ClientLevel level, double x, double y, double z, int delay, TextureAtlasSprite sprite) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprite);
         this.quadSize = 0.85F;
         this.delay = delay;
