--- a/net/minecraft/client/particle/ShriekParticle.java
+++ b/net/minecraft/client/particle/ShriekParticle.java
@@ -14,7 +_,7 @@
     private static final float MAGICAL_X_ROT = 1.0472F;
     private int delay;
 
-    ShriekParticle(ClientLevel p_233976_, double p_233977_, double p_233978_, double p_233979_, int p_233980_) {
+    protected ShriekParticle(ClientLevel p_233976_, double p_233977_, double p_233978_, double p_233979_, int p_233980_) {
         super(p_233976_, p_233977_, p_233978_, p_233979_, 0.0, 0.0, 0.0);
         this.quadSize = 0.85F;
         this.delay = p_233980_;
