--- a/net/minecraft/client/particle/SimpleVerticalParticle.java
+++ b/net/minecraft/client/particle/SimpleVerticalParticle.java
@@ -9,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SimpleVerticalParticle extends SingleQuadParticle {
-    public SimpleVerticalParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite, boolean upwards) {
+    private SimpleVerticalParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite, boolean upwards) {
         super(level, x, y, z, xa, ya, za, sprite);
         this.xd = xa;
         this.zd = za;
