--- a/net/minecraft/client/particle/SimpleAnimatedParticle.java
+++ b/net/minecraft/client/particle/SimpleAnimatedParticle.java
@@ -12,7 +_,7 @@
     private float fadeB;
     private boolean hasFade;
 
-    public SimpleAnimatedParticle(ClientLevel level, double x, double y, double z, SpriteSet sprites, float gravity) {
+    protected SimpleAnimatedParticle(ClientLevel level, double x, double y, double z, SpriteSet sprites, float gravity) {
         super(level, x, y, z, sprites.first());
         this.friction = 0.91F;
         this.gravity = gravity;
