--- a/net/minecraft/client/particle/SonicBoomParticle.java
+++ b/net/minecraft/client/particle/SonicBoomParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SonicBoomParticle extends HugeExplosionParticle {
-    public SonicBoomParticle(ClientLevel level, double x, double y, double z, double size, SpriteSet sprites) {
+    protected SonicBoomParticle(ClientLevel level, double x, double y, double z, double size, SpriteSet sprites) {
         super(level, x, y, z, size, sprites);
         this.lifetime = 16;
         this.quadSize = 1.5F;
