--- a/net/minecraft/client/particle/SoulParticle.java
+++ b/net/minecraft/client/particle/SoulParticle.java
@@ -12,7 +_,7 @@
     private final SpriteSet sprites;
     protected boolean isGlowing;
 
-    public SoulParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, SpriteSet sprites) {
+    private SoulParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, SpriteSet sprites) {
         super(level, x, y, z, xd, yd, zd, sprites.first());
         this.sprites = sprites;
         this.scale(1.5F);
