--- a/net/minecraft/client/particle/SoulParticle.java
+++ b/net/minecraft/client/particle/SoulParticle.java
@@ -10,7 +_,7 @@
     private final SpriteSet sprites;
     protected boolean isGlowing;
 
-    SoulParticle(
+    protected SoulParticle(
         ClientLevel p_107717_, double p_107718_, double p_107719_, double p_107720_, double p_107721_, double p_107722_, double p_107723_, SpriteSet p_107724_
     ) {
         super(p_107717_, p_107718_, p_107719_, p_107720_, p_107721_, p_107722_, p_107723_);
