--- a/net/minecraft/client/particle/SpellParticle.java
+++ b/net/minecraft/client/particle/SpellParticle.java
@@ -17,7 +_,7 @@
     private final SpriteSet sprites;
     private float originalAlpha = 1.0F;
 
-    public SpellParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
+    private SpellParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
         super(level, x, y, z, 0.5 - RANDOM.nextDouble(), ya, 0.5 - RANDOM.nextDouble(), sprites.first());
         this.friction = 0.96F;
         this.gravity = -0.1F;
