--- a/net/minecraft/client/particle/SpellParticle.java
+++ b/net/minecraft/client/particle/SpellParticle.java
@@ -16,7 +_,7 @@
     private final SpriteSet sprites;
     private float originalAlpha = 1.0F;
 
-    SpellParticle(
+    protected SpellParticle(
         ClientLevel p_107762_, double p_107763_, double p_107764_, double p_107765_, double p_107766_, double p_107767_, double p_107768_, SpriteSet p_107769_
     ) {
         super(p_107762_, p_107763_, p_107764_, p_107765_, 0.5 - RANDOM.nextDouble(), p_107767_, 0.5 - RANDOM.nextDouble());
