--- a/net/minecraft/client/particle/SpitParticle.java
+++ b/net/minecraft/client/particle/SpitParticle.java
@@ -8,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SpitParticle extends ExplodeParticle {
-    public SpitParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
+    private SpitParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
         super(level, x, y, z, xa, ya, za, sprites);
         this.gravity = 0.5F;
     }
