--- a/net/minecraft/client/particle/SpitParticle.java
+++ b/net/minecraft/client/particle/SpitParticle.java
@@ -7,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SpitParticle extends ExplodeParticle {
-    SpitParticle(
+    protected SpitParticle(
         ClientLevel p_107888_, double p_107889_, double p_107890_, double p_107891_, double p_107892_, double p_107893_, double p_107894_, SpriteSet p_107895_
     ) {
         super(p_107888_, p_107889_, p_107890_, p_107891_, p_107892_, p_107893_, p_107894_, p_107895_);
