--- a/net/minecraft/client/particle/SplashParticle.java
+++ b/net/minecraft/client/particle/SplashParticle.java
@@ -7,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SplashParticle extends WaterDropParticle {
-    SplashParticle(ClientLevel p_107929_, double p_107930_, double p_107931_, double p_107932_, double p_107933_, double p_107934_, double p_107935_) {
+    protected SplashParticle(ClientLevel p_107929_, double p_107930_, double p_107931_, double p_107932_, double p_107933_, double p_107934_, double p_107935_) {
         super(p_107929_, p_107930_, p_107931_, p_107932_);
         this.gravity = 0.04F;
         if (p_107934_ == 0.0 && (p_107933_ != 0.0 || p_107935_ != 0.0)) {
