--- a/net/minecraft/client/particle/SplashParticle.java
+++ b/net/minecraft/client/particle/SplashParticle.java
@@ -9,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SplashParticle extends WaterDropParticle {
-    public SplashParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
+    private SplashParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
         super(level, x, y, z, sprite);
         this.gravity = 0.04F;
         if (ya == 0.0 && (xa != 0.0 || za != 0.0)) {
