--- a/net/minecraft/client/particle/SquidInkParticle.java
+++ b/net/minecraft/client/particle/SquidInkParticle.java
@@ -10,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SquidInkParticle extends SimpleAnimatedParticle {
-    public SquidInkParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, int color, SpriteSet sprites) {
+    private SquidInkParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, int color, SpriteSet sprites) {
         super(level, x, y, z, sprites, 0.0F);
         this.friction = 0.92F;
         this.quadSize = 0.5F;
