--- a/net/minecraft/client/particle/SulfurBubbleParticle.java
+++ b/net/minecraft/client/particle/SulfurBubbleParticle.java
@@ -20,7 +_,7 @@
     private final float sizeStart;
     private double yPrev;
 
-    public SulfurBubbleParticle(ClientLevel level, double x, double y, double z, double xa, double za, TextureAtlasSprite sprite) {
+    private SulfurBubbleParticle(ClientLevel level, double x, double y, double z, double xa, double za, TextureAtlasSprite sprite) {
         super(level, x, y, z, sprite);
         this.gravity = -0.04F;
         this.friction = 0.85F;
