--- a/net/minecraft/client/particle/SuspendedTownParticle.java
+++ b/net/minecraft/client/particle/SuspendedTownParticle.java
@@ -7,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SuspendedTownParticle extends TextureSheetParticle {
-    SuspendedTownParticle(ClientLevel p_108104_, double p_108105_, double p_108106_, double p_108107_, double p_108108_, double p_108109_, double p_108110_) {
+    protected SuspendedTownParticle(ClientLevel p_108104_, double p_108105_, double p_108106_, double p_108107_, double p_108108_, double p_108109_, double p_108110_) {
         super(p_108104_, p_108105_, p_108106_, p_108107_, p_108108_, p_108109_, p_108110_);
         float f = this.random.nextFloat() * 0.1F + 0.2F;
         this.rCol = f;
