--- a/net/minecraft/client/particle/SuspendedParticle.java
+++ b/net/minecraft/client/particle/SuspendedParticle.java
@@ -11,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SuspendedParticle extends TextureSheetParticle {
-    SuspendedParticle(ClientLevel p_172403_, SpriteSet p_172404_, double p_172405_, double p_172406_, double p_172407_) {
+    protected SuspendedParticle(ClientLevel p_172403_, SpriteSet p_172404_, double p_172405_, double p_172406_, double p_172407_) {
         super(p_172403_, p_172405_, p_172406_ - 0.125, p_172407_);
         this.setSize(0.01F, 0.01F);
         this.pickSprite(p_172404_);
