--- a/net/minecraft/client/particle/SuspendedParticle.java
+++ b/net/minecraft/client/particle/SuspendedParticle.java
@@ -12,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SuspendedParticle extends SingleQuadParticle {
-    public SuspendedParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
+    private SuspendedParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
         super(level, x, y - 0.125, z, sprite);
         this.setSize(0.01F, 0.01F);
         this.quadSize = this.quadSize * (this.random.nextFloat() * 0.6F + 0.2F);
@@ -22,7 +_,7 @@
         this.gravity = 0.0F;
     }
 
-    public SuspendedParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
+    private SuspendedParticle(ClientLevel level, double x, double y, double z, double xd, double yd, double zd, TextureAtlasSprite sprite) {
         super(level, x, y - 0.125, z, xd, yd, zd, sprite);
         this.setSize(0.01F, 0.01F);
         this.quadSize = this.quadSize * (this.random.nextFloat() * 0.6F + 0.6F);
