--- a/net/minecraft/client/particle/SuspendedTownParticle.java
+++ b/net/minecraft/client/particle/SuspendedTownParticle.java
@@ -9,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class SuspendedTownParticle extends SingleQuadParticle {
-    public SuspendedTownParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
+    private SuspendedTownParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, TextureAtlasSprite sprite) {
         super(level, x, y, z, xa, ya, za, sprite);
         float br = this.random.nextFloat() * 0.1F + 0.2F;
         this.rCol = br;
