--- a/net/minecraft/client/particle/TotemParticle.java
+++ b/net/minecraft/client/particle/TotemParticle.java
@@ -7,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class TotemParticle extends SimpleAnimatedParticle {
-    TotemParticle(
+    protected TotemParticle(
         ClientLevel p_108346_, double p_108347_, double p_108348_, double p_108349_, double p_108350_, double p_108351_, double p_108352_, SpriteSet p_108353_
     ) {
         super(p_108346_, p_108347_, p_108348_, p_108349_, p_108353_, 1.25F);
