--- a/net/minecraft/client/particle/TrackingEmitter.java
+++ b/net/minecraft/client/particle/TrackingEmitter.java
@@ -22,7 +_,7 @@
         this(level, entity, particleType, lifeTime, entity.getDeltaMovement());
     }
 
-    public TrackingEmitter(ClientLevel level, Entity entity, ParticleOptions particleType, int lifeTime, Vec3 movement) {
+    private TrackingEmitter(ClientLevel level, Entity entity, ParticleOptions particleType, int lifeTime, Vec3 movement) {
         super(level, entity.getX(), entity.getY(0.5), entity.getZ(), movement.x, movement.y, movement.z);
         this.entity = entity;
         this.lifeTime = lifeTime;
