--- a/net/minecraft/client/particle/TrailParticle.java
+++ b/net/minecraft/client/particle/TrailParticle.java
@@ -14,7 +_,7 @@
 public class TrailParticle extends SingleQuadParticle {
     private final Vec3 target;
 
-    public TrailParticle(
+    private TrailParticle(
         ClientLevel level, double x, double y, double z, double xAux, double yAux, double zAux, Vec3 target, int color, TextureAtlasSprite sprite
     ) {
         super(level, x, y, z, xAux, yAux, zAux, sprite);
