--- a/net/minecraft/client/particle/TrialSpawnerDetectionParticle.java
+++ b/net/minecraft/client/particle/TrialSpawnerDetectionParticle.java
@@ -13,7 +_,7 @@
     private final SpriteSet sprites;
     private static final int BASE_LIFETIME = 8;
 
-    public TrialSpawnerDetectionParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
+    protected TrialSpawnerDetectionParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprites.first());
         this.sprites = sprites;
         this.friction = 0.96F;
