--- a/net/minecraft/client/particle/VibrationSignalParticle.java
+++ b/net/minecraft/client/particle/VibrationSignalParticle.java
@@ -20,7 +_,7 @@
     private float pitch;
     private float pitchO;
 
-    VibrationSignalParticle(ClientLevel p_234105_, double p_234106_, double p_234107_, double p_234108_, PositionSource p_234109_, int p_234110_) {
+    protected VibrationSignalParticle(ClientLevel p_234105_, double p_234106_, double p_234107_, double p_234108_, PositionSource p_234109_, int p_234110_) {
         super(p_234105_, p_234106_, p_234107_, p_234108_, 0.0, 0.0, 0.0);
         this.quadSize = 0.3F;
         this.target = p_234109_;
@@ -76,6 +_,7 @@
                 this.x = Mth.lerp(d0, this.x, vec3.x());
                 this.y = Mth.lerp(d0, this.y, vec3.y());
                 this.z = Mth.lerp(d0, this.z, vec3.z());
+                this.setPos(this.x, this.y, this.z); // FORGE: Update the particle's bounding box
                 double d1 = this.x - vec3.x();
                 double d2 = this.y - vec3.y();
                 double d3 = this.z - vec3.z();
