--- a/net/minecraft/client/particle/WakeParticle.java
+++ b/net/minecraft/client/particle/WakeParticle.java
@@ -10,7 +_,7 @@
 public class WakeParticle extends SingleQuadParticle {
     private final SpriteSet sprites;
 
-    public WakeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
+    private WakeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, SpriteSet sprites) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprites.first());
         this.sprites = sprites;
         this.xd *= 0.3F;
