--- a/net/minecraft/client/particle/WakeParticle.java
+++ b/net/minecraft/client/particle/WakeParticle.java
@@ -9,7 +_,7 @@
 public class WakeParticle extends TextureSheetParticle {
     private final SpriteSet sprites;
 
-    WakeParticle(
+    protected WakeParticle(
         ClientLevel p_108407_, double p_108408_, double p_108409_, double p_108410_, double p_108411_, double p_108412_, double p_108413_, SpriteSet p_108414_
     ) {
         super(p_108407_, p_108408_, p_108409_, p_108410_, 0.0, 0.0, 0.0);
