--- a/net/minecraft/client/particle/WaterCurrentDownParticle.java
+++ b/net/minecraft/client/particle/WaterCurrentDownParticle.java
@@ -12,7 +_,7 @@
 public class WaterCurrentDownParticle extends TextureSheetParticle {
     private float angle;
 
-    WaterCurrentDownParticle(ClientLevel p_108450_, double p_108451_, double p_108452_, double p_108453_) {
+    protected WaterCurrentDownParticle(ClientLevel p_108450_, double p_108451_, double p_108452_, double p_108453_) {
         super(p_108450_, p_108451_, p_108452_, p_108453_);
         this.lifetime = (int)(Math.random() * 60.0) + 30;
         this.hasPhysics = false;
