--- a/net/minecraft/client/particle/WaterDropParticle.java
+++ b/net/minecraft/client/particle/WaterDropParticle.java
@@ -11,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class WaterDropParticle extends SingleQuadParticle {
-    public WaterDropParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
+    protected WaterDropParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
         super(level, x, y, z, 0.0, 0.0, 0.0, sprite);
         this.xd *= 0.3F;
         this.yd = this.random.nextFloat() * 0.2F + 0.1F;
