--- a/net/minecraft/client/particle/WaterCurrentDownParticle.java
+++ b/net/minecraft/client/particle/WaterCurrentDownParticle.java
@@ -14,7 +_,7 @@
 public class WaterCurrentDownParticle extends SingleQuadParticle {
     private float angle;
 
-    public WaterCurrentDownParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
+    private WaterCurrentDownParticle(ClientLevel level, double x, double y, double z, TextureAtlasSprite sprite) {
         super(level, x, y, z, sprite);
         this.lifetime = (int)(this.random.nextFloat() * 60.0F) + 30;
         this.hasPhysics = false;
