--- a/net/minecraft/client/particle/WhiteAshParticle.java
+++ b/net/minecraft/client/particle/WhiteAshParticle.java
@@ -11,7 +_,7 @@
 public class WhiteAshParticle extends BaseAshSmokeParticle {
     private static final int COLOR_RGB24 = 12235202;
 
-    public WhiteAshParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
+    protected WhiteAshParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
         super(level, x, y, z, 0.1F, -0.1F, 0.1F, xa, ya, za, scale, sprites, 0.0F, 20, 0.0125F, false);
         this.rCol = ARGB.red(12235202) / 255.0F;
         this.gCol = ARGB.green(12235202) / 255.0F;
