--- a/net/minecraft/client/particle/WhiteSmokeParticle.java
+++ b/net/minecraft/client/particle/WhiteSmokeParticle.java
@@ -10,7 +_,7 @@
 public class WhiteSmokeParticle extends BaseAshSmokeParticle {
     private static final int COLOR_RGB24 = 12235202;
 
-    public WhiteSmokeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
+    protected WhiteSmokeParticle(ClientLevel level, double x, double y, double z, double xa, double ya, double za, float scale, SpriteSet sprites) {
         super(level, x, y, z, 0.1F, 0.1F, 0.1F, xa, ya, za, scale, sprites, 0.3F, 8, -0.1F, true);
         this.rCol = 0.7294118F;
         this.gCol = 0.69411767F;
