--- a/com/mojang/renderpearl/api/textures/GpuSampler.java
+++ b/com/mojang/renderpearl/api/textures/GpuSampler.java
@@ -1,12 +_,11 @@
 package com.mojang.renderpearl.api.textures;
 
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.util.OptionalDouble;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface GpuSampler extends UncheckedAutoCloseable {
+public interface GpuSampler extends AutoCloseable {
     AddressMode getAddressModeU();
 
     AddressMode getAddressModeV();
@@ -20,4 +_,7 @@
     OptionalDouble getMaxLod();
 
     boolean isClosed();
+
+    @Override
+    void close();
 }
