--- a/com/mojang/renderpearl/api/textures/GpuTexture.java
+++ b/com/mojang/renderpearl/api/textures/GpuTexture.java
@@ -1,7 +_,6 @@
 package com.mojang.renderpearl.api.textures;
 
 import com.mojang.renderpearl.api.GpuFormat;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
@@ -10,7 +_,7 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface GpuTexture extends UncheckedAutoCloseable {
+public interface GpuTexture extends AutoCloseable {
     int USAGE_COPY_DST = 1;
     int USAGE_COPY_SRC = 2;
     int USAGE_TEXTURE_BINDING = 4;
@@ -30,6 +_,9 @@
     @GpuTexture.Usage int usage();
 
     String getLabel();
+
+    @Override
+    void close();
 
     boolean isClosed();
 
