--- a/net/minecraft/util/parsing/packrat/commands/Grammar.java
+++ b/net/minecraft/util/parsing/packrat/commands/Grammar.java
@@ -7,7 +_,6 @@
 import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
-import java.util.function.Consumer;
 import java.util.stream.Collectors;
 import net.minecraft.commands.SharedSuggestionProvider;
 import net.minecraft.util.parsing.packrat.Atom;
