--- a/com/mojang/renderpearl/util/TextureViewAndSampler.java
+++ /dev/null
@@ -1,10 +_,0 @@
-package com.mojang.renderpearl.util;
-
-import com.mojang.renderpearl.api.textures.GpuSampler;
-import com.mojang.renderpearl.api.textures.GpuTextureView;
-import net.neoforged.api.distmarker.Dist;
-import net.neoforged.api.distmarker.OnlyIn;
-
-@OnlyIn(Dist.CLIENT)
-public record TextureViewAndSampler(GpuTextureView view, GpuSampler sampler) {
-}
