--- a/net/minecraft/world/level/dimension/DimensionDefaults.java
+++ b/net/minecraft/world/level/dimension/DimensionDefaults.java
@@ -10,12 +_,6 @@
     public static final int OVERWORLD_GENERATION_HEIGHT = 384;
     public static final int OVERWORLD_LOGICAL_HEIGHT = 384;
     public static final float OVERWORLD_CLOUD_HEIGHT = 192.33F;
-    public static final int OVERWORLD_CELL_SIZE_XZ = 4;
-    public static final int OVERWORLD_CELL_SIZE_Y = 8;
-    public static final int NETHER_CELL_SIZE_XZ = 4;
-    public static final int NETHER_CELL_SIZE_Y = 8;
-    public static final int END_CELL_SIZE_XZ = 8;
-    public static final int END_CELL_SIZE_Y = 4;
     public static final int NETHER_MIN_Y = 0;
     public static final int NETHER_LEVEL_HEIGHT = 256;
     public static final int NETHER_GENERATION_HEIGHT = 128;
