--- a/net/minecraft/client/multiplayer/ClientAdvancements.java
+++ b/net/minecraft/client/multiplayer/ClientAdvancements.java
@@ -1,6 +_,5 @@
 package net.minecraft.client.multiplayer;
 
-import com.google.common.collect.Iterables;
 import com.mojang.logging.LogUtils;
 import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
 import java.util.Map;
@@ -41,34 +_,28 @@
         if (packet.shouldReset()) {
             this.tree.clear();
             this.progress.clear();
-            if (this.listener != null) {
-                this.listener.onAdvancementsCleared();
-            }
-        }
-
-        this.tree.remove(packet.removed());
-        this.tree.addAll(Iterables.transform(packet.added(), ClientboundUpdateAdvancementsPacket.PositionedAdvancement::advancement));
-
-        for (ClientboundUpdateAdvancementsPacket.PositionedAdvancement positionedAdvancement : packet.added()) {
-            AdvancementNode node = this.tree.get(positionedAdvancement.advancement());
-            if (node != null) {
-                node.setLocation(positionedAdvancement.x(), positionedAdvancement.y());
-            }
-        }
-
-        for (Entry<Identifier, AdvancementProgress> entry : packet.progress().entrySet()) {
+        }
+
+        this.tree.remove(packet.getRemoved());
+        this.tree.addAll(packet.getAdded());
+
+        for (Entry<Identifier, AdvancementProgress> entry : packet.getProgress().entrySet()) {
             AdvancementNode node = this.tree.get(entry.getKey());
             if (node != null) {
                 AdvancementProgress progress = entry.getValue();
                 progress.update(node.advancement().requirements());
                 this.progress.put(node.holder(), progress);
+                if (this.listener != null) {
+                    this.listener.onUpdateAdvancementProgress(node, progress);
+                }
+
                 if (!packet.shouldReset() && progress.isDone()) {
                     if (this.minecraft.level != null) {
                         this.telemetryManager.onAdvancementDone(this.minecraft.level, node.holder());
                     }
 
                     Optional<DisplayInfo> display = node.advancement().display();
-                    if (packet.showAdvancements() && display.isPresent() && display.get().showToast()) {
+                    if (packet.shouldShowAdvancements() && display.isPresent() && display.get().shouldShowToast()) {
                         this.minecraft.gui.toastManager().addToast(new AdvancementToast(node.holder()));
                     }
                 }
@@ -76,17 +_,9 @@
                 LOGGER.warn("Server informed client about progress for unknown advancement {}", entry.getKey());
             }
         }
-
-        if (this.listener != null) {
-            this.listener.onAdvancementsUpdated();
-        }
-    }
-
-    public Map<AdvancementHolder, AdvancementProgress> progress() {
-        return this.progress;
-    }
-
-    public AdvancementTree tree() {
+    }
+
+    public AdvancementTree getTree() {
         return this.tree;
     }
 
@@ -106,8 +_,14 @@
 
     public void setListener(ClientAdvancements.@Nullable Listener listener) {
         this.listener = listener;
+        this.tree.setListener(listener);
         if (listener != null) {
-            listener.onAdvancementsUpdated();
+            this.progress.forEach((holder, progress) -> {
+                AdvancementNode node = this.tree.get(holder);
+                if (node != null) {
+                    listener.onUpdateAdvancementProgress(node, progress);
+                }
+            });
             listener.onSelectedTabChanged(this.selectedTab);
         }
     }
@@ -117,10 +_,8 @@
         return node != null ? node.holder() : null;
     }
 
-    public interface Listener {
-        void onAdvancementsUpdated();
-
-        void onAdvancementsCleared();
+    public interface Listener extends AdvancementTree.Listener {
+        void onUpdateAdvancementProgress(AdvancementNode advancement, AdvancementProgress progress);
 
         void onSelectedTabChanged(@Nullable AdvancementHolder selectedTab);
     }
