--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -260,6 +_,7 @@
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.SignatureValidator;
 import net.minecraft.util.debug.DebugValueAccess;
+import net.minecraft.util.thread.ReentrantBlockableEventLoop;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.SimpleContainer;
@@ -329,6 +_,7 @@
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
+import net.neoforged.neoforge.network.connection.ConnectionType;
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
@@ -2529,6 +_,16 @@
 
     public Connection getConnection() {
         return this.connection;
+    }
+
+    @Override
+    public ReentrantBlockableEventLoop<?> getMainThreadEventLoop() {
+        return null;
+    }
+
+    @Override
+    public ConnectionType getConnectionType() {
+        return null;
     }
 
     @Override
