--- a/net/minecraft/client/multiplayer/ClientConfigurationPacketListenerImpl.java
+++ b/net/minecraft/client/multiplayer/ClientConfigurationPacketListenerImpl.java
@@ -88,7 +_,7 @@
     @Override
     public void handleUpdateTags(ClientboundUpdateTagsPacket packet) {
         PacketUtils.ensureRunningOnSameThread(packet, this, this.minecraft.packetProcessor());
-        this.registryDataCollector.appendTags(packet.tags());
+        this.registryDataCollector.appendTags(packet.getTags());
     }
 
     @Override
