--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -400,7 +_,7 @@
     private int serverChunkRadius = 3;
     private int serverSimulationDistance = 3;
     private final RandomSource random = RandomSource.createThreadSafe();
-    public CommandDispatcher<ClientSuggestionProvider> commands = new CommandDispatcher<>();
+    private CommandDispatcher<ClientSuggestionProvider> commands = new CommandDispatcher<>();
     private ClientRecipeContainer recipes = new ClientRecipeContainer(Map.of(), SelectableRecipe.SingleInputSet.empty());
     private Set<ResourceKey<Level>> levels;
     private final RegistryAccess.Frozen registryAccess;
