--- a/net/minecraft/client/multiplayer/ClientLevel.java
+++ b/net/minecraft/client/multiplayer/ClientLevel.java
@@ -62,6 +_,7 @@
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.server.level.BlockDestructionProgress;
 import net.minecraft.server.level.ParticleStatus;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
@@ -114,6 +_,7 @@
 import net.minecraft.world.level.border.WorldBorder;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.dimension.DimensionType;
+import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.entity.EntityTickList;
 import net.minecraft.world.level.entity.LevelCallback;
 import net.minecraft.world.level.entity.LevelEntityGetter;
@@ -522,6 +_,11 @@
         return true;
     }
 
+    @Override
+    public ResourceKey<LevelStem> getTypeKey() {
+        return null;
+    }
+
     public int getEntityCount() {
         return this.entityStorage.count();
     }
@@ -987,6 +_,11 @@
 
     @Override
     public void gameEvent(Holder<GameEvent> gameEvent, Vec3 pos, GameEvent.Context context) {
+    }
+
+    @Override
+    public ServerLevel getMinecraftWorld() {
+        return null;
     }
 
     protected Map<MapId, MapItemSavedData> getAllMapData() {
