--- a/net/minecraft/client/multiplayer/ClientConfigurationPacketListenerImpl.java
+++ b/net/minecraft/client/multiplayer/ClientConfigurationPacketListenerImpl.java
@@ -15,6 +_,7 @@
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.network.Connection;
 import net.minecraft.network.DisconnectionDetails;
+import net.minecraft.network.PacketProcessor;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.TickablePacketListener;
 import net.minecraft.network.chat.Component;
@@ -40,6 +_,7 @@
 import net.minecraft.world.flag.FeatureFlags;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
+import net.neoforged.neoforge.network.connection.ConnectionType;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -216,5 +_,20 @@
                 ClientConfigurationPacketListenerImpl.LOGGER.warn("Commands are not supported in configuration phase, trying to run '{}'", command);
             }
         };
+    }
+
+    @Override
+    public Connection getConnection() {
+        return null;
+    }
+
+    @Override
+    public PacketProcessor getPacketProcessor() {
+        return null;
+    }
+
+    @Override
+    public ConnectionType getConnectionType() {
+        return null;
     }
 }
