--- a/net/minecraft/client/multiplayer/ClientSuggestionProvider.java
+++ b/net/minecraft/client/multiplayer/ClientSuggestionProvider.java
@@ -11,7 +_,6 @@
 import java.util.Locale;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
-import java.util.function.Predicate;
 import java.util.stream.Stream;
 import net.minecraft.client.Minecraft;
 import net.minecraft.commands.SharedSuggestionProvider;
@@ -98,15 +_,11 @@
     }
 
     @Override
-    public <E> CompletableFuture<Suggestions> suggestRegistryElements(
-        ResourceKey<? extends Registry<E>> key,
-        SharedSuggestionProvider.ElementSuggestionType elements,
-        SuggestionsBuilder builder,
-        CommandContext<?> context,
-        Predicate<E> filter
+    public CompletableFuture<Suggestions> suggestRegistryElements(
+        ResourceKey<? extends Registry<?>> key, SharedSuggestionProvider.ElementSuggestionType elements, SuggestionsBuilder builder, CommandContext<?> context
     ) {
         return this.registryAccess().lookup(key).map(registry -> {
-            this.suggestRegistryElements(registry, elements, builder, filter);
+            this.suggestRegistryElements(registry, elements, builder);
             return builder.buildFuture();
         }).orElseGet(() -> this.customSuggestion(context));
     }
