--- a/net/minecraft/client/multiplayer/SessionSearchTrees.java
+++ b/net/minecraft/client/multiplayer/SessionSearchTrees.java
@@ -31,8 +_,8 @@
 @OnlyIn(Dist.CLIENT)
 public class SessionSearchTrees {
     private static final SessionSearchTrees.Key RECIPE_COLLECTIONS = new SessionSearchTrees.Key();
-    public static final SessionSearchTrees.Key CREATIVE_NAMES = new SessionSearchTrees.Key();
-    public static final SessionSearchTrees.Key CREATIVE_TAGS = new SessionSearchTrees.Key();
+    private static final SessionSearchTrees.Key CREATIVE_NAMES = new SessionSearchTrees.Key();
+    private static final SessionSearchTrees.Key CREATIVE_TAGS = new SessionSearchTrees.Key();
     private CompletableFuture<SearchTree<ItemStack>> creativeByNameSearch = CompletableFuture.completedFuture(SearchTree.empty());
     private CompletableFuture<SearchTree<ItemStack>> creativeByTagSearch = CompletableFuture.completedFuture(SearchTree.empty());
     private CompletableFuture<SearchTree<RecipeCollection>> recipeSearch = CompletableFuture.completedFuture(SearchTree.empty());
@@ -129,6 +_,6 @@
         return this.creativeByNameSearch.join();
     }
 
-    public static class Key {
+    private static class Key {
     }
 }
