--- a/net/minecraft/world/entity/ai/Brain.java
+++ b/net/minecraft/world/entity/ai/Brain.java
@@ -195,7 +_,7 @@
     public <U> Optional<U> getMemory(MemoryModuleType<U> p_21953_) {
         Optional<? extends ExpirableValue<?>> optional = this.memories.get(p_21953_);
         if (optional == null) {
-            throw new IllegalStateException("Unregistered memory fetched: " + p_21953_);
+            return Optional.empty();
         } else {
             return (Optional<U>)optional.map(ExpirableValue::getValue);
         }
