--- a/net/minecraft/world/level/gameevent/GameEvent.java
+++ b/net/minecraft/world/level/gameevent/GameEvent.java
@@ -5,8 +_,8 @@
 import net.minecraft.core.Registry;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.resources.Identifier;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
@@ -75,7 +_,7 @@
     public static final Holder.Reference<GameEvent> RESONATE_14 = register("resonate_14");
     public static final Holder.Reference<GameEvent> RESONATE_15 = register("resonate_15");
     public static final int DEFAULT_NOTIFICATION_RADIUS = 16;
-    public static final Codec<Holder<GameEvent>> CODEC = RegistryCodecs.holder(Registries.GAME_EVENT);
+    public static final Codec<Holder<GameEvent>> CODEC = RegistryFixedCodec.create(Registries.GAME_EVENT);
 
     public static Holder<GameEvent> bootstrap(Registry<GameEvent> registry) {
         return BLOCK_ACTIVATE;
@@ -86,7 +_,7 @@
     }
 
     private static Holder.Reference<GameEvent> register(String name, int notificationRadius) {
-        return Registry.registerForHolder(BuiltInRegistries.GAME_EVENT, Identifier.withDefaultNamespace(name), new GameEvent(notificationRadius));
+        return io.papermc.paper.registry.PaperRegistryListenerManager.INSTANCE.registerForHolderWithListeners(BuiltInRegistries.GAME_EVENT, Identifier.withDefaultNamespace(name), new GameEvent(notificationRadius)); // Paper - run with listeners
     }
 
     public record Context(@Nullable Entity sourceEntity, @Nullable BlockState affectedState) {
