--- a/net/minecraft/world/entity/animal/goat/GoatAi.java
+++ b/net/minecraft/world/entity/animal/goat/GoatAi.java
@@ -101,8 +_,7 @@
                 )
             ),
             ImmutableSet.of(
-                Pair.of(MemoryModuleType.RAM_COOLDOWN_TICKS, MemoryStatus.VALUE_PRESENT),
-                Pair.of(MemoryModuleType.LONG_JUMP_COOLDOWN_TICKS, MemoryStatus.VALUE_PRESENT)
+                Pair.of(MemoryModuleType.RAM_TARGET, MemoryStatus.VALUE_ABSENT), Pair.of(MemoryModuleType.LONG_JUMP_MID_JUMP, MemoryStatus.VALUE_ABSENT)
             )
         );
     }
