--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -286,7 +286,7 @@
         }
     }
 
-    public void applyBiomeDecoration(WorldGenLevel worldgenlevel, ChunkAccess chunkaccess, StructureManager structuremanager) {
+    public void addVanillaDecorations(WorldGenLevel worldgenlevel, ChunkAccess chunkaccess, StructureManager structuremanager) { // CraftBukkit
         ChunkPos chunkpos = chunkaccess.getPos();
 
         if (!SharedConstants.debugVoidTerrain(chunkpos)) {
@@ -305,7 +305,7 @@
                 ChunkAccess chunkaccess1 = worldgenlevel.getChunk(chunkpos1.x, chunkpos1.z);
 
                 for (LevelChunkSection levelchunksection : chunkaccess1.getSections()) {
-                    PalettedContainerRO palettedcontainerro = levelchunksection.getBiomes();
+                    PalettedContainerRO<Holder<Biome>> palettedcontainerro = levelchunksection.getBiomes(); // CraftBukkit - decompile error
 
                     Objects.requireNonNull(set);
                     palettedcontainerro.getAll(set::add);
@@ -323,7 +323,7 @@
                     int i1 = 0;
 
                     if (structuremanager.shouldGenerateStructures()) {
-                        for (Structure structure : (List) map.getOrDefault(l, Collections.emptyList())) {
+                        for (Structure structure : map.getOrDefault(l, Collections.emptyList())) { // CraftBukkit - decompile error
                             worldgenrandom.setFeatureSeed(i, i1, l);
                             Supplier<String> supplier = () -> {
                                 Optional optional = registry.getResourceKey(structure).map(Object::toString);
@@ -413,6 +413,33 @@
         }
     }
 
+   // CraftBukkit start
+    public void applyBiomeDecoration(WorldGenLevel worldgenlevel, ChunkAccess chunkaccess, StructureManager structuremanager) {
+        applyBiomeDecoration(worldgenlevel, chunkaccess, structuremanager, true);
+    }
+
+    public void applyBiomeDecoration(WorldGenLevel worldgenlevel, ChunkAccess chunkaccess, StructureManager structuremanager, boolean vanilla) {
+        if (vanilla) {
+            addVanillaDecorations(worldgenlevel, chunkaccess, structuremanager);
+        }
+
+        org.bukkit.World world = worldgenlevel.getMinecraftWorld().getWorld();
+        // only call when a populator is present (prevents unnecessary entity conversion)
+        if (!world.getPopulators().isEmpty()) {
+            org.bukkit.craftbukkit.generator.CraftLimitedRegion limitedRegion = new org.bukkit.craftbukkit.generator.CraftLimitedRegion(worldgenlevel, chunkaccess.getPos());
+            int x = chunkaccess.getPos().x;
+            int z = chunkaccess.getPos().z;
+            for (org.bukkit.generator.BlockPopulator populator : world.getPopulators()) {
+                WorldgenRandom worldgenrandom = new WorldgenRandom(new net.minecraft.world.level.levelgen.LegacyRandomSource(worldgenlevel.getSeed()));
+                worldgenrandom.setDecorationSeed(worldgenlevel.getSeed(), x, z);
+                populator.populate(world, new org.bukkit.craftbukkit.util.RandomSourceWrapper.RandomWrapper(worldgenrandom), x, z, limitedRegion);
+            }
+            limitedRegion.saveEntities();
+            limitedRegion.breakLink();
+        }
+    }
+    // CraftBukkit end
+
     private static BoundingBox getWritableArea(ChunkAccess chunkaccess) {
         ChunkPos chunkpos = chunkaccess.getPos();
         int i = chunkpos.getMinBlockX();
@@ -541,6 +568,14 @@
         StructureStart structurestart = structure.generate(structureset_structureselectionentry.structure(), resourcekey, registryaccess, this, this.biomeSource, randomstate, structuretemplatemanager, i, chunkpos, j, chunkaccess, predicate);
 
         if (structurestart.isValid()) {
+            // CraftBukkit start
+            BoundingBox box = structurestart.getBoundingBox();
+            org.bukkit.event.world.AsyncStructureSpawnEvent event = new org.bukkit.event.world.AsyncStructureSpawnEvent(structuremanager.level.getMinecraftWorld().getWorld(), org.bukkit.craftbukkit.generator.structure.CraftStructure.minecraftToBukkit(structure), new org.bukkit.util.BoundingBox(box.minX(), box.minY(), box.minZ(), box.maxX(), box.maxY(), box.maxZ()), chunkpos.x, chunkpos.z);
+            org.bukkit.Bukkit.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return true;
+            }
+            // CraftBukkit end
             structuremanager.setStartForStructure(sectionpos, structure, structurestart, chunkaccess);
             return true;
         } else {
