--- a/net/minecraft/world/food/FoodData.java
+++ b/net/minecraft/world/food/FoodData.java
@@ -13,6 +_,13 @@
     private int tickTimer;
     private int lastFoodLevel = 20;
 
+    // CraftBukkit start
+    public Player entityhuman = null;
+    public int saturatedRegenRate = 10;
+    public int unsaturatedRegenRate = 80;
+    public int starvationRate = 80;
+    // CraftBukkit end
+
     public FoodData() {
         this.saturationLevel = 5.0F;
     }
