--- a/net/minecraft/world/food/FoodData.java
+++ b/net/minecraft/world/food/FoodData.java
@@ -7,9 +_,9 @@
 import net.minecraft.world.level.GameRules;
 
 public class FoodData {
-    public int foodLevel = 20;
-    public float saturationLevel;
-    public float exhaustionLevel;
+    private int foodLevel = 20;
+    private float saturationLevel;
+    private float exhaustionLevel;
     private int tickTimer;
     private int lastFoodLevel = 20;
 
