--- a/net/minecraft/client/resources/model/geometry/UnbakedGeometry.java
+++ b/net/minecraft/client/resources/model/geometry/UnbakedGeometry.java
@@ -7,11 +_,9 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
-import net.neoforged.neoforge.client.extensions.UnbakedGeometryExtension;
-
 @FunctionalInterface
 @OnlyIn(Dist.CLIENT)
-public interface UnbakedGeometry extends UnbakedGeometryExtension {
+public interface UnbakedGeometry {
     UnbakedGeometry EMPTY = (var0, var1, var2, var3) -> QuadCollection.EMPTY;
 
     QuadCollection bake(TextureSlots textureSlots, ModelBaker modelBaker, ModelState modelState, ModelDebugName name);
