--- a/net/minecraft/client/input/InputQuirks.java
+++ b/net/minecraft/client/input/InputQuirks.java
@@ -6,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class InputQuirks {
-    public static final boolean ON_OSX = Util.getPlatform() == Util.OS.OSX;
+    private static final boolean ON_OSX = Util.getPlatform() == Util.OS.OSX;
     public static final boolean REPLACE_CTRL_KEY_WITH_CMD_KEY = ON_OSX;
     public static final int EDIT_SHORTCUT_KEY_MODIFIER = REPLACE_CTRL_KEY_WITH_CMD_KEY ? 8 : 2;
     public static final boolean SIMULATE_RIGHT_CLICK_WITH_LONG_LEFT_CLICK = ON_OSX;
