--- a/net/minecraft/client/input/InputWithModifiers.java
+++ b/net/minecraft/client/input/InputWithModifiers.java
@@ -10,6 +_,8 @@
 
 @OnlyIn(Dist.CLIENT)
 public interface InputWithModifiers {
+    int NOT_DIGIT = -1;
+
     @InputConstants.Value int input();
 
     default int shortcutKey() {
@@ -48,6 +_,11 @@
 
     default boolean isCycleFocus() {
         return this.input() == 43;
+    }
+
+    default int getDigit() {
+        int value = this.input() - 39;
+        return value >= 0 && value <= 9 ? value : -1;
     }
 
     default boolean hasAltDown() {
