--- a/net/minecraft/client/renderer/entity/layers/CrossedArmsItemLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/CrossedArmsItemLayer.java
@@ -30,9 +_,9 @@
 
     protected void applyTranslation(S state, PoseStack poseStack) {
         this.getParentModel().translateToArms(state, poseStack);
-        poseStack.rotate(Axis.XP, 0.75F);
+        poseStack.mulPose(Axis.XP.rotation(0.75F));
         poseStack.scale(1.07F, 1.07F, 1.07F);
         poseStack.translate(0.0F, 0.13F, -0.34F);
-        poseStack.rotate(Axis.XP, (float) Math.PI);
+        poseStack.mulPose(Axis.XP.rotation((float) Math.PI));
     }
 }
