--- a/net/minecraft/client/renderer/entity/layers/EquipmentLayerRenderer.java
+++ b/net/minecraft/client/renderer/entity/layers/EquipmentLayerRenderer.java
@@ -102,7 +_,7 @@
         }
     }
 
-    public static int getColorForLayer(EquipmentClientInfo.Layer layer, int dyeColor) {
+    private static int getColorForLayer(EquipmentClientInfo.Layer layer, int dyeColor) {
         Optional<EquipmentClientInfo.Dyeable> dyeable = layer.dyeable();
         if (dyeable.isPresent()) {
             int colorWhenUndyed = dyeable.get().colorWhenUndyed().map(ARGB::opaque).orElse(0);
