--- a/net/minecraft/client/renderer/entity/layers/FoxHeldItemLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/FoxHeldItemLayer.java
@@ -29,9 +_,9 @@
                 poseStack.scale(0.75F, 0.75F, 0.75F);
             }
 
-            poseStack.rotate(Axis.ZP, state.headRollAngle);
-            poseStack.rotateDegrees(Axis.YP, yRot);
-            poseStack.rotateDegrees(Axis.XP, xRot);
+            poseStack.mulPose(Axis.ZP.rotation(state.headRollAngle));
+            poseStack.mulPose(Axis.YP.rotationDegrees(yRot));
+            poseStack.mulPose(Axis.XP.rotationDegrees(xRot));
             if (state.isBaby) {
                 if (sleeping) {
                     poseStack.translate(0.4F, 0.26F, 0.15F);
@@ -44,9 +_,9 @@
                 poseStack.translate(0.06F, 0.27F, -0.5F);
             }
 
-            poseStack.rotateDegrees(Axis.XP, 90.0F);
+            poseStack.mulPose(Axis.XP.rotationDegrees(90.0F));
             if (sleeping) {
-                poseStack.rotateDegrees(Axis.ZP, 90.0F);
+                poseStack.mulPose(Axis.ZP.rotationDegrees(90.0F));
             }
 
             item.submit(poseStack, submitNodeCollector, lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
