--- a/net/minecraft/client/renderer/entity/layers/IronGolemFlowerLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/IronGolemFlowerLayer.java
@@ -26,7 +_,7 @@
             poseStack.translate(0.5F, 0.5F, 0.5F);
             float s = 0.5F;
             poseStack.scale(0.5F, 0.5F, 0.5F);
-            poseStack.rotateDegrees(Axis.XP, -90.0F);
+            poseStack.mulPose(Axis.XP.rotationDegrees(-90.0F));
             poseStack.translate(-0.5F, -0.5F, -0.5F);
             state.flowerBlock.submit(poseStack, submitNodeCollector, lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
             poseStack.popPose();
