--- a/net/minecraft/client/renderer/entity/layers/ItemInHandLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/ItemInHandLayer.java
@@ -34,16 +_,14 @@
         if (!item.isEmpty()) {
             poseStack.pushPose();
             this.getParentModel().translateToHand(state, arm, poseStack);
-            poseStack.rotateDegrees(Axis.XP, -90.0F);
-            poseStack.rotateDegrees(Axis.YP, 180.0F);
+            poseStack.mulPose(Axis.XP.rotationDegrees(-90.0F));
+            poseStack.mulPose(Axis.YP.rotationDegrees(180.0F));
             boolean isLeftHand = arm == HumanoidArm.LEFT;
             float offsetX = this.useBabyOffset(state) ? 0.0F : 1.0F;
             float offsetY = this.useBabyOffset(state) ? 1.0F : 2.0F;
             float offsetZ = this.useBabyOffset(state) ? -4.5F : -10.0F;
             poseStack.translate((isLeftHand ? -1 : 1) * offsetX / 16.0F, offsetY / 16.0F, offsetZ / 16.0F);
-            if (state.currentSwing != null
-                && state.currentSwing.hand().asArm(state.mainArm) == arm
-                && state.currentSwing.animation().type() == SwingAnimationType.STAB) {
+            if (state.attackTime > 0.0F && state.attackArm == arm && state.swingAnimationType == SwingAnimationType.STAB) {
                 SpearAnimations.thirdPersonAttackItem(state, poseStack);
             }
 
