--- a/net/minecraft/client/renderer/entity/layers/PlayerItemInHandLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/PlayerItemInHandLayer.java
@@ -11,6 +_,7 @@
 import net.minecraft.client.renderer.item.ItemStackRenderState;
 import net.minecraft.client.renderer.texture.OverlayTexture;
 import net.minecraft.util.Mth;
+import net.minecraft.world.InteractionHand;
 import net.minecraft.world.entity.HumanoidArm;
 import net.minecraft.world.item.ItemStack;
 import net.neoforged.api.distmarker.Dist;
@@ -29,7 +_,8 @@
         S state, ItemStackRenderState item, ItemStack itemStack, HumanoidArm arm, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, int lightCoords
     ) {
         if (!item.isEmpty()) {
-            if (state.isUsingItem && state.useItemHand.asArm(state.mainArm) == arm && state.swingAnimation < 1.0E-5F && !state.heldOnHead.isEmpty()) {
+            InteractionHand currentHand = arm == state.mainArm ? InteractionHand.MAIN_HAND : InteractionHand.OFF_HAND;
+            if (state.isUsingItem && state.useItemHand == currentHand && state.attackTime < 1.0E-5F && !state.heldOnHead.isEmpty()) {
                 this.renderItemHeldToEye(state, arm, poseStack, submitNodeCollector, lightCoords);
             } else {
                 super.submitArmWithItem(state, item, itemStack, arm, poseStack, submitNodeCollector, lightCoords);
