--- a/net/minecraft/client/renderer/entity/layers/SnowGolemHeadLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/SnowGolemHeadLayer.java
@@ -23,7 +_,7 @@
                 this.getParentModel().getHead().translateAndRotate(poseStack);
                 float s = 0.625F;
                 poseStack.translate(0.0F, -0.34375F, 0.0F);
-                poseStack.rotateDegrees(Axis.YP, 180.0F);
+                poseStack.mulPose(Axis.YP.rotationDegrees(180.0F));
                 poseStack.scale(0.625F, -0.625F, -0.625F);
                 int overlayCoords = LivingEntityRenderer.getOverlayCoords(state, 0.0F);
                 poseStack.translate(-0.5F, -0.5F, -0.5F);
