--- a/net/minecraft/client/renderer/entity/layers/StuckInBodyLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/StuckInBodyLayer.java
@@ -44,8 +_,8 @@
         float directionXZ = Mth.sqrt(directionX * directionX + directionZ * directionZ);
         float yRot = (float)(Math.atan2(directionX, directionZ) * 180.0F / (float)Math.PI);
         float xRot = (float)(Math.atan2(directionY, directionXZ) * 180.0F / (float)Math.PI);
-        poseStack.rotateDegrees(Axis.YP, yRot - 90.0F);
-        poseStack.rotateDegrees(Axis.ZP, xRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(yRot - 90.0F));
+        poseStack.mulPose(Axis.ZP.rotationDegrees(xRot));
         submitNodeCollector.submitModel(this.model, this.modelState, poseStack, this.texture, lightCoords, OverlayTexture.NO_OVERLAY, outlineColor);
     }
 
