--- a/net/minecraft/client/renderer/entity/layers/SulfurCubeInnerLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/SulfurCubeInnerLayer.java
@@ -39,7 +_,7 @@
                 : LivingEntityRenderer.getOverlayCoords(state, 0.0F);
             if (!state.containedBlock.isEmpty()) {
                 poseStack.pushPose();
-                poseStack.rotateDegrees(Axis.XP, 180.0F);
+                poseStack.mulPose(Axis.XP.rotationDegrees(180.0F));
                 if (state.isBaby) {
                     poseStack.scale(0.5F, 0.5F, 0.5F);
                 }
