--- a/net/minecraft/client/renderer/entity/layers/WitchItemLayer.java
+++ b/net/minecraft/client/renderer/entity/layers/WitchItemLayer.java
@@ -20,10 +_,10 @@
             this.getParentModel().translateToHead(poseStack);
             this.getParentModel().getNose().translateAndRotate(poseStack);
             poseStack.translate(0.0625F, 0.25F, 0.0F);
-            poseStack.rotateDegrees(Axis.ZP, 180.0F);
-            poseStack.rotateDegrees(Axis.XP, 140.0F);
-            poseStack.rotateDegrees(Axis.ZP, 10.0F);
-            poseStack.rotateDegrees(Axis.XP, 180.0F);
+            poseStack.mulPose(Axis.ZP.rotationDegrees(180.0F));
+            poseStack.mulPose(Axis.XP.rotationDegrees(140.0F));
+            poseStack.mulPose(Axis.ZP.rotationDegrees(10.0F));
+            poseStack.mulPose(Axis.XP.rotationDegrees(180.0F));
         } else {
             super.applyTranslation(state, poseStack);
         }
