--- a/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/GroundPathNavigation.java
@@ -3,11 +_,11 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.SectionPos;
-import net.minecraft.tags.FluidTags;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.Mob;
 import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.pathfinder.Node;
@@ -41,7 +_,7 @@
     }
 
     @Override
-    public Path createPath(BlockPos pos, int reachRange) {
+    public Path createPath(BlockPos pos, @org.jspecify.annotations.Nullable Entity entity, int reachRange) { // Paper - EntityPathfindEvent
         LevelChunk chunk = this.level.getChunkSource().getChunkNow(SectionPos.blockToSectionCoord(pos.getX()), SectionPos.blockToSectionCoord(pos.getZ()));
         if (chunk == null) {
             return null;
@@ -51,7 +_,7 @@
             pos = this.findSurfacePosition(chunk, pos, reachRange);
         }
 
-        return super.createPath(pos, reachRange);
+        return super.createPath(pos, entity, reachRange); // Paper - EntityPathfindEvent
     }
 
     private BlockPos findSurfacePosition(LevelChunk chunk, BlockPos pos, int reachRange) {
@@ -90,16 +_,16 @@
 
     @Override
     public Path createPath(Entity target, int reachRange) {
-        return this.createPath(target.blockPosition(), reachRange);
+        return this.createPath(target.blockPosition(), target, reachRange); // Paper - EntityPathfindEvent
     }
 
     private int getSurfaceY() {
-        if (this.mob.isInFloatableFluid() && this.canFloat()) {
+        if (this.mob.isInWater() && this.canFloat()) {
             int surface = this.mob.getBlockY();
             BlockState state = this.level.getBlockState(BlockPos.containing(this.mob.getX(), surface, this.mob.getZ()));
             int steps = 0;
 
-            while (state.getFluidState().is(FluidTags.ENTITY_FLOATABLE)) {
+            while (state.is(Blocks.WATER)) {
                 state = this.level.getBlockState(BlockPos.containing(this.mob.getX(), ++surface, this.mob.getZ()));
                 if (++steps > 16) {
                     return this.mob.getBlockY();
