--- /dev/null
+++ b/net/minecraft/client/renderer/rendertype/OutputTarget.java
@@ -1,0 +_,47 @@
+package net.minecraft.client.renderer.rendertype;
+
+import com.mojang.blaze3d.pipeline.RenderTarget;
+import java.util.function.Supplier;
+import net.minecraft.client.Minecraft;
+import net.minecraft.client.renderer.LevelRenderer;
+import net.neoforged.api.distmarker.Dist;
+import net.neoforged.api.distmarker.OnlyIn;
+import org.jspecify.annotations.Nullable;
+
+@OnlyIn(Dist.CLIENT)
+public class OutputTarget {
+    private final String name;
+    private final Supplier<@Nullable RenderTarget> renderTargetSupplier;
+    public static final OutputTarget MAIN_TARGET = new OutputTarget("main_target", () -> Minecraft.getInstance().gameRenderer.mainRenderTarget());
+    public static final OutputTarget OUTLINE_TARGET = new OutputTarget("outline_target", () -> Minecraft.getInstance().levelRenderer.entityOutlineTarget());
+    public static final OutputTarget TERRAIN_DEPTH_TARGET = new OutputTarget(
+        "terrain_depth_target", () -> Minecraft.getInstance().levelRenderer.terrainDepthTarget()
+    );
+    public static final OutputTarget DEPTH_BOUNDS_TARGET = new OutputTarget(
+        "depth_bounds_target", () -> Minecraft.getInstance().levelRenderer.depthBoundsTarget()
+    );
+    public static final OutputTarget[] TRANSMITTANCE_TARGETS = new OutputTarget[LevelRenderer.OIT_TRANSMITTANCE_TARGET_COUNT];
+    public static final OutputTarget ACCUMULATE_TARGET = new OutputTarget("accumulate_target", () -> Minecraft.getInstance().levelRenderer.accumulateTarget());
+
+    public OutputTarget(String name, Supplier<@Nullable RenderTarget> renderTargetSupplier) {
+        this.name = name;
+        this.renderTargetSupplier = renderTargetSupplier;
+    }
+
+    public RenderTarget getRenderTarget() {
+        RenderTarget preferredTarget = this.renderTargetSupplier.get();
+        return preferredTarget != null ? preferredTarget : Minecraft.getInstance().gameRenderer.mainRenderTarget();
+    }
+
+    @Override
+    public String toString() {
+        return "OutputTarget[" + this.name + "]";
+    }
+
+    static {
+        for (int i = 0; i < LevelRenderer.OIT_TRANSMITTANCE_TARGET_COUNT; i++) {
+            int finalI = i;
+            TRANSMITTANCE_TARGETS[i] = new OutputTarget("transmittance_target_" + i, () -> Minecraft.getInstance().levelRenderer.transmittanceTarget(finalI));
+        }
+    }
+}
