--- a/net/minecraft/client/renderer/rendertype/PreparedRenderType.java
+++ b/net/minecraft/client/renderer/rendertype/PreparedRenderType.java
@@ -50,7 +_,7 @@
         renderPass.setVertexBuffer(0, info.vertexBuffer().slice());
 
         for (PreparedRenderType.Texture texture : this.textures) {
-            renderPass.setUniform(texture.name, texture.textureView, texture.sampler);
+            renderPass.bindTexture(texture.name, texture.textureView, texture.sampler);
         }
 
         renderPass.setIndexBuffer(indexBuffer, info.indexType());
