--- a/net/minecraft/client/renderer/rendertype/RenderSetup.java
+++ b/net/minecraft/client/renderer/rendertype/RenderSetup.java
@@ -28,7 +_,6 @@
     final Map<String, RenderSetup.TextureBinding> textures;
     final TextureTransform textureTransform;
     final RenderSetup.OutlineProperty outlineProperty;
-    final @Nullable String outlineTextureName;
     final boolean useLightmap;
     final boolean useOverlay;
     final boolean affectsCrumbling;
@@ -45,7 +_,6 @@
         LayeringTransform layeringTransform,
         TextureTransform textureTransform,
         RenderSetup.OutlineProperty outlineProperty,
-        @Nullable String outlineTextureName,
         boolean affectsCrumbling,
         boolean sortOnUpload,
         boolean forceSolidModelPhase
@@ -57,7 +_,6 @@
         this.useLightmap = useLightmap;
         this.useOverlay = useOverlay;
         this.outlineProperty = outlineProperty;
-        this.outlineTextureName = outlineTextureName;
         this.layeringTransform = layeringTransform;
         this.affectsCrumbling = affectsCrumbling;
         this.sortOnUpload = sortOnUpload;
@@ -139,7 +_,6 @@
         private boolean affectsCrumbling = false;
         private boolean sortOnUpload = false;
         private RenderSetup.OutlineProperty outlineProperty = RenderSetup.OutlineProperty.NONE;
-        private @Nullable String outlineTextureName;
         private final Map<String, RenderSetup.TextureBinding> textures = new HashMap<>();
         private boolean forceSolidModelPhase;
 
@@ -189,13 +_,6 @@
 
         public RenderSetup.RenderSetupBuilder setOutline(RenderSetup.OutlineProperty outlineProperty) {
             this.outlineProperty = outlineProperty;
-            this.outlineTextureName = null;
-            return this;
-        }
-
-        public RenderSetup.RenderSetupBuilder setOutline(RenderSetup.OutlineProperty outlineProperty, String outlineTextureName) {
-            this.outlineProperty = outlineProperty;
-            this.outlineTextureName = outlineTextureName;
             return this;
         }
 
@@ -219,7 +_,6 @@
                 this.layeringTransform,
                 this.textureTransform,
                 this.outlineProperty,
-                this.outlineTextureName,
                 this.affectsCrumbling,
                 this.sortOnUpload,
                 this.forceSolidModelPhase
