--- a/net/minecraft/core/registries/BuiltInRegistries.java
+++ b/net/minecraft/core/registries/BuiltInRegistries.java
@@ -60,7 +_,6 @@
 import net.minecraft.stats.StatType;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.Util;
-import net.minecraft.util.context.ContextKeySet;
 import net.minecraft.util.debug.DebugSubscription;
 import net.minecraft.util.debug.DebugSubscriptions;
 import net.minecraft.util.valueproviders.FloatProvider;
@@ -125,11 +_,12 @@
 import net.minecraft.world.level.gameevent.PositionSourceType;
 import net.minecraft.world.level.gamerules.GameRule;
 import net.minecraft.world.level.gamerules.GameRules;
+import net.minecraft.world.level.levelgen.DensityFunction;
+import net.minecraft.world.level.levelgen.DensityFunctions;
+import net.minecraft.world.level.levelgen.SurfaceRules;
 import net.minecraft.world.level.levelgen.blockpredicates.BlockPredicateType;
 import net.minecraft.world.level.levelgen.carver.WorldCarver;
 import net.minecraft.world.level.levelgen.carver.WorldCarverTypes;
-import net.minecraft.world.level.levelgen.densityfunction.DensityFunction;
-import net.minecraft.world.level.levelgen.densityfunction.DensityFunctions;
 import net.minecraft.world.level.levelgen.feature.Feature;
 import net.minecraft.world.level.levelgen.feature.FeatureTypes;
 import net.minecraft.world.level.levelgen.feature.featuresize.FeatureSizeType;
@@ -140,9 +_,6 @@
 import net.minecraft.world.level.levelgen.feature.treedecorators.TreeDecoratorType;
 import net.minecraft.world.level.levelgen.feature.trunkplacers.TrunkPlacerType;
 import net.minecraft.world.level.levelgen.heightproviders.HeightProviderType;
-import net.minecraft.world.level.levelgen.material.MaterialRules;
-import net.minecraft.world.level.levelgen.material.condition.MaterialCondition;
-import net.minecraft.world.level.levelgen.material.rule.MaterialRule;
 import net.minecraft.world.level.levelgen.placement.PlacementModifier;
 import net.minecraft.world.level.levelgen.placement.PlacementModifierTypes;
 import net.minecraft.world.level.levelgen.structure.StructureType;
@@ -165,15 +_,12 @@
 import net.minecraft.world.level.storage.loot.entries.LootPoolEntryContainer;
 import net.minecraft.world.level.storage.loot.functions.LootItemFunction;
 import net.minecraft.world.level.storage.loot.functions.LootItemFunctions;
-import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
 import net.minecraft.world.level.storage.loot.predicates.LootItemConditionTypes;
 import net.minecraft.world.level.storage.loot.providers.nbt.NbtProvider;
 import net.minecraft.world.level.storage.loot.providers.nbt.NbtProviders;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProvider;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProviderTypes;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviderTypes;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviderTypes;
 import net.minecraft.world.level.storage.loot.providers.score.ScoreboardNameProvider;
 import net.minecraft.world.level.storage.loot.providers.score.ScoreboardNameProviders;
 import org.slf4j.Logger;
@@ -186,44 +_,46 @@
     );
     public static final DataComponentInitializers DATA_COMPONENT_INITIALIZERS = new DataComponentInitializers();
     public static final DefaultedRegistry<GameEvent> GAME_EVENT = registerDefaulted(Registries.GAME_EVENT, "step", GameEvent::bootstrap);
-    public static final Registry<SoundEvent> SOUND_EVENT = registerSimple(Registries.SOUND_EVENT, var0 -> SoundEvents.ITEM_PICKUP);
-    public static final DefaultedRegistry<Fluid> FLUID = registerDefaultedWithIntrusiveHolders(Registries.FLUID, "empty", var0 -> Fluids.EMPTY);
+    public static final Registry<SoundEvent> SOUND_EVENT = registerSimple(Registries.SOUND_EVENT, registry -> SoundEvents.ITEM_PICKUP);
+    public static final DefaultedRegistry<Fluid> FLUID = registerDefaultedWithIntrusiveHolders(Registries.FLUID, "empty", registry -> Fluids.EMPTY);
     public static final Registry<MobEffect> MOB_EFFECT = registerSimple(Registries.MOB_EFFECT, MobEffects::bootstrap);
-    public static final DefaultedRegistry<Block> BLOCK = registerDefaultedWithIntrusiveHolders(Registries.BLOCK, "air", var0 -> Blocks.AIR);
+    public static final DefaultedRegistry<Block> BLOCK = registerDefaultedWithIntrusiveHolders(Registries.BLOCK, "air", registry -> Blocks.AIR);
     public static final Registry<DebugSubscription<?>> DEBUG_SUBSCRIPTION = registerSimple(Registries.DEBUG_SUBSCRIPTION, DebugSubscriptions::bootstrap);
     public static final DefaultedRegistry<EntityType<?>> ENTITY_TYPE = registerDefaultedWithIntrusiveHolders(
-        Registries.ENTITY_TYPE, "pig", var0 -> EntityTypes.PIG
+        Registries.ENTITY_TYPE, "pig", registry -> EntityTypes.PIG
     );
-    public static final DefaultedRegistry<Item> ITEM = registerDefaultedWithIntrusiveHolders(Registries.ITEM, "air", var0 -> Items.AIR);
+    public static final DefaultedRegistry<Item> ITEM = registerDefaultedWithIntrusiveHolders(Registries.ITEM, "air", registry -> Items.AIR);
     public static final Registry<Potion> POTION = registerSimple(Registries.POTION, Potions::bootstrap);
-    public static final Registry<ParticleType<?>> PARTICLE_TYPE = registerSimple(Registries.PARTICLE_TYPE, var0 -> ParticleTypes.BLOCK);
+    public static final Registry<ParticleType<?>> PARTICLE_TYPE = registerSimple(Registries.PARTICLE_TYPE, registry -> ParticleTypes.BLOCK);
     public static final Registry<BlockEntityType<?>> BLOCK_ENTITY_TYPE = registerSimpleWithIntrusiveHolders(
-        Registries.BLOCK_ENTITY_TYPE, var0 -> BlockEntityTypes.FURNACE
+        Registries.BLOCK_ENTITY_TYPE, registry -> BlockEntityTypes.FURNACE
     );
-    public static final Registry<Identifier> CUSTOM_STAT = registerSimple(Registries.CUSTOM_STAT, var0 -> Stats.JUMP);
-    public static final DefaultedRegistry<ChunkStatus> CHUNK_STATUS = registerDefaulted(Registries.CHUNK_STATUS, "empty", var0 -> ChunkStatus.EMPTY);
-    public static final Registry<RuleTestType<?>> RULE_TEST = registerSimple(Registries.RULE_TEST, var0 -> RuleTestType.ALWAYS_TRUE_TEST);
+    public static final Registry<Identifier> CUSTOM_STAT = registerSimple(Registries.CUSTOM_STAT, registry -> Stats.JUMP);
+    public static final DefaultedRegistry<ChunkStatus> CHUNK_STATUS = registerDefaulted(Registries.CHUNK_STATUS, "empty", registry -> ChunkStatus.EMPTY);
+    public static final Registry<RuleTestType<?>> RULE_TEST = registerSimple(Registries.RULE_TEST, registry -> RuleTestType.ALWAYS_TRUE_TEST);
     public static final Registry<RuleBlockEntityModifierType<?>> RULE_BLOCK_ENTITY_MODIFIER = registerSimple(
-        Registries.RULE_BLOCK_ENTITY_MODIFIER, var0 -> RuleBlockEntityModifierType.PASSTHROUGH
+        Registries.RULE_BLOCK_ENTITY_MODIFIER, registry -> RuleBlockEntityModifierType.PASSTHROUGH
     );
-    public static final Registry<PosRuleTestType<?>> POS_RULE_TEST = registerSimple(Registries.POS_RULE_TEST, var0 -> PosRuleTestType.ALWAYS_TRUE_TEST);
-    public static final Registry<MenuType<?>> MENU = registerSimple(Registries.MENU, var0 -> MenuType.ANVIL);
-    public static final Registry<RecipeType<?>> RECIPE_TYPE = registerSimple(Registries.RECIPE_TYPE, var0 -> RecipeType.CRAFTING);
+    public static final Registry<PosRuleTestType<?>> POS_RULE_TEST = registerSimple(Registries.POS_RULE_TEST, registry -> PosRuleTestType.ALWAYS_TRUE_TEST);
+    public static final Registry<MenuType<?>> MENU = registerSimple(Registries.MENU, registry -> MenuType.ANVIL);
+    public static final Registry<RecipeType<?>> RECIPE_TYPE = registerSimple(Registries.RECIPE_TYPE, registry -> RecipeType.CRAFTING);
     public static final Registry<RecipeSerializer<?>> RECIPE_SERIALIZER = registerSimple(Registries.RECIPE_SERIALIZER, RecipeSerializers::bootstrap);
     public static final Registry<Attribute> ATTRIBUTE = registerSimple(Registries.ATTRIBUTE, Attributes::bootstrap);
-    public static final Registry<PositionSourceType<?>> POSITION_SOURCE_TYPE = registerSimple(Registries.POSITION_SOURCE_TYPE, var0 -> PositionSourceType.BLOCK);
+    public static final Registry<PositionSourceType<?>> POSITION_SOURCE_TYPE = registerSimple(
+        Registries.POSITION_SOURCE_TYPE, registry -> PositionSourceType.BLOCK
+    );
     public static final Registry<ArgumentTypeInfo<?, ?>> COMMAND_ARGUMENT_TYPE = registerSimple(Registries.COMMAND_ARGUMENT_TYPE, ArgumentTypeInfos::bootstrap);
-    public static final Registry<StatType<?>> STAT_TYPE = registerSimple(Registries.STAT_TYPE, var0 -> Stats.ITEM_USED);
+    public static final Registry<StatType<?>> STAT_TYPE = registerSimple(Registries.STAT_TYPE, registry -> Stats.ITEM_USED);
     public static final DefaultedRegistry<VillagerType> VILLAGER_TYPE = registerDefaulted(Registries.VILLAGER_TYPE, "plains", VillagerType::bootstrap);
     public static final DefaultedRegistry<VillagerProfession> VILLAGER_PROFESSION = registerDefaulted(
         Registries.VILLAGER_PROFESSION, "none", VillagerProfession::bootstrap
     );
     public static final Registry<PoiType> POINT_OF_INTEREST_TYPE = registerSimple(Registries.POINT_OF_INTEREST_TYPE, PoiTypes::bootstrap);
     public static final DefaultedRegistry<MemoryModuleType<?>> MEMORY_MODULE_TYPE = registerDefaulted(
-        Registries.MEMORY_MODULE_TYPE, "dummy", var0 -> MemoryModuleType.DUMMY
+        Registries.MEMORY_MODULE_TYPE, "dummy", registry -> MemoryModuleType.DUMMY
     );
-    public static final DefaultedRegistry<SensorType<?>> SENSOR_TYPE = registerDefaulted(Registries.SENSOR_TYPE, "dummy", var0 -> SensorType.DUMMY);
-    public static final Registry<Activity> ACTIVITY = registerSimple(Registries.ACTIVITY, var0 -> Activity.IDLE);
+    public static final DefaultedRegistry<SensorType<?>> SENSOR_TYPE = registerDefaulted(Registries.SENSOR_TYPE, "dummy", registry -> SensorType.DUMMY);
+    public static final Registry<Activity> ACTIVITY = registerSimple(Registries.ACTIVITY, registry -> Activity.IDLE);
     public static final Registry<MapCodec<? extends LootPoolEntryContainer>> LOOT_POOL_ENTRY_TYPE = registerSimple(
         Registries.LOOT_POOL_ENTRY_TYPE, LootPoolEntries::bootstrap
     );
@@ -233,11 +_,8 @@
     public static final Registry<MapCodec<? extends LootItemCondition>> LOOT_CONDITION_TYPE = registerSimple(
         Registries.LOOT_CONDITION_TYPE, LootItemConditionTypes::bootstrap
     );
-    public static final Registry<MapCodec<? extends ContextFloatProvider>> CONTEXT_FLOAT_PROVIDER_TYPE = registerSimple(
-        Registries.CONTEXT_FLOAT_PROVIDER_TYPE, ContextFloatProviderTypes::bootstrap
-    );
-    public static final Registry<MapCodec<? extends ContextIntProvider>> CONTEXT_INT_PROVIDER_TYPE = registerSimple(
-        Registries.CONTEXT_INT_PROVIDER_TYPE, ContextIntProviderTypes::bootstrap
+    public static final Registry<MapCodec<? extends NumberProvider>> LOOT_NUMBER_PROVIDER_TYPE = registerSimple(
+        Registries.LOOT_NUMBER_PROVIDER_TYPE, NumberProviderTypes::bootstrap
     );
     public static final Registry<MapCodec<? extends NbtProvider>> LOOT_NBT_PROVIDER_TYPE = registerSimple(
         Registries.LOOT_NBT_PROVIDER_TYPE, NbtProviders::bootstrap
@@ -250,16 +_,20 @@
     );
     public static final Registry<MapCodec<? extends IntProvider>> INT_PROVIDER_TYPE = registerSimple(Registries.INT_PROVIDER_TYPE, IntProviders::bootstrap);
     public static final Registry<HeightProviderType<?>> HEIGHT_PROVIDER_TYPE = registerSimple(
-        Registries.HEIGHT_PROVIDER_TYPE, var0 -> HeightProviderType.CONSTANT
-    );
-    public static final Registry<BlockPredicateType<?>> BLOCK_PREDICATE_TYPE = registerSimple(Registries.BLOCK_PREDICATE_TYPE, var0 -> BlockPredicateType.NOT);
+        Registries.HEIGHT_PROVIDER_TYPE, registry -> HeightProviderType.CONSTANT
+    );
+    public static final Registry<BlockPredicateType<?>> BLOCK_PREDICATE_TYPE = registerSimple(
+        Registries.BLOCK_PREDICATE_TYPE, registry -> BlockPredicateType.NOT
+    );
     public static final Registry<MapCodec<? extends WorldCarver>> CARVER_TYPE = registerSimple(Registries.CARVER_TYPE, WorldCarverTypes::bootstrap);
     public static final Registry<MapCodec<? extends Feature>> FEATURE_TYPE = registerSimple(Registries.FEATURE_TYPE, FeatureTypes::bootstrap);
     public static final Registry<MapCodec<? extends StructurePlacement>> STRUCTURE_PLACEMENT = registerSimple(
         Registries.STRUCTURE_PLACEMENT, StructurePlacements::bootstrap
     );
-    public static final Registry<StructurePieceType> STRUCTURE_PIECE = registerSimple(Registries.STRUCTURE_PIECE, var0 -> StructurePieceType.MINE_SHAFT_ROOM);
-    public static final Registry<StructureType<?>> STRUCTURE_TYPE = registerSimple(Registries.STRUCTURE_TYPE, var0 -> StructureType.JIGSAW);
+    public static final Registry<StructurePieceType> STRUCTURE_PIECE = registerSimple(
+        Registries.STRUCTURE_PIECE, registry -> StructurePieceType.MINE_SHAFT_ROOM
+    );
+    public static final Registry<StructureType<?>> STRUCTURE_TYPE = registerSimple(Registries.STRUCTURE_TYPE, registry -> StructureType.JIGSAW);
     public static final Registry<MapCodec<? extends PlacementModifier>> PLACEMENT_MODIFIER_TYPE = registerSimple(
         Registries.PLACEMENT_MODIFIER_TYPE, PlacementModifierTypes::bootstrap
     );
@@ -267,25 +_,27 @@
         Registries.BLOCK_STATE_PROVIDER_TYPE, BlockStateProviderTypes::bootstrap
     );
     public static final Registry<FoliagePlacerType<?>> FOLIAGE_PLACER_TYPE = registerSimple(
-        Registries.FOLIAGE_PLACER_TYPE, var0 -> FoliagePlacerType.BLOB_FOLIAGE_PLACER
+        Registries.FOLIAGE_PLACER_TYPE, registry -> FoliagePlacerType.BLOB_FOLIAGE_PLACER
     );
     public static final Registry<TrunkPlacerType<?>> TRUNK_PLACER_TYPE = registerSimple(
-        Registries.TRUNK_PLACER_TYPE, var0 -> TrunkPlacerType.STRAIGHT_TRUNK_PLACER
-    );
-    public static final Registry<RootPlacerType<?>> ROOT_PLACER_TYPE = registerSimple(Registries.ROOT_PLACER_TYPE, var0 -> RootPlacerType.MANGROVE_ROOT_PLACER);
+        Registries.TRUNK_PLACER_TYPE, registry -> TrunkPlacerType.STRAIGHT_TRUNK_PLACER
+    );
+    public static final Registry<RootPlacerType<?>> ROOT_PLACER_TYPE = registerSimple(
+        Registries.ROOT_PLACER_TYPE, registry -> RootPlacerType.MANGROVE_ROOT_PLACER
+    );
     public static final Registry<TreeDecoratorType<?>> TREE_DECORATOR_TYPE = registerSimple(
-        Registries.TREE_DECORATOR_TYPE, var0 -> TreeDecoratorType.LEAVE_VINE
+        Registries.TREE_DECORATOR_TYPE, registry -> TreeDecoratorType.LEAVE_VINE
     );
     public static final Registry<FeatureSizeType<?>> FEATURE_SIZE_TYPE = registerSimple(
-        Registries.FEATURE_SIZE_TYPE, var0 -> FeatureSizeType.TWO_LAYERS_FEATURE_SIZE
+        Registries.FEATURE_SIZE_TYPE, registry -> FeatureSizeType.TWO_LAYERS_FEATURE_SIZE
     );
     public static final Registry<MapCodec<? extends BiomeSource>> BIOME_SOURCE = registerSimple(Registries.BIOME_SOURCE, BiomeSources::bootstrap);
     public static final Registry<MapCodec<? extends ChunkGenerator>> CHUNK_GENERATOR = registerSimple(Registries.CHUNK_GENERATOR, ChunkGenerators::bootstrap);
-    public static final Registry<MapCodec<? extends MaterialCondition>> MATERIAL_CONDITION_TYPE = registerSimple(
-        Registries.MATERIAL_CONDITION_TYPE, MaterialRules::bootstrapConditions
+    public static final Registry<MapCodec<? extends SurfaceRules.ConditionSource>> MATERIAL_CONDITION_TYPE = registerSimple(
+        Registries.MATERIAL_CONDITION_TYPE, SurfaceRules.ConditionSource::bootstrap
     );
-    public static final Registry<MapCodec<? extends MaterialRule>> MATERIAL_RULE_TYPE = registerSimple(
-        Registries.MATERIAL_RULE_TYPE, MaterialRules::bootstrapRules
+    public static final Registry<MapCodec<? extends SurfaceRules.RuleSource>> MATERIAL_RULE_TYPE = registerSimple(
+        Registries.MATERIAL_RULE_TYPE, SurfaceRules.RuleSource::bootstrap
     );
     public static final Registry<MapCodec<? extends DensityFunction>> DENSITY_FUNCTION_TYPE = registerSimple(
         Registries.DENSITY_FUNCTION_TYPE, DensityFunctions::bootstrap
@@ -294,7 +_,7 @@
         Registries.STRUCTURE_PROCESSOR, StructureProcessorTypes::bootstrap
     );
     public static final Registry<StructurePoolElementType<?>> STRUCTURE_POOL_ELEMENT = registerSimple(
-        Registries.STRUCTURE_POOL_ELEMENT, var0 -> StructurePoolElementType.EMPTY
+        Registries.STRUCTURE_POOL_ELEMENT, registry -> StructurePoolElementType.EMPTY
     );
     public static final Registry<MapCodec<? extends PoolAliasBinding>> POOL_ALIAS_BINDING_TYPE = registerSimple(
         Registries.POOL_ALIAS_BINDING, PoolAliasBindings::bootstrap
@@ -330,15 +_,15 @@
         Registries.ENCHANTMENT_PROVIDER_TYPE, EnchantmentProviderTypes::bootstrap
     );
     public static final Registry<ConsumeEffect.Type<?>> CONSUME_EFFECT_TYPE = registerSimple(
-        Registries.CONSUME_EFFECT_TYPE, var0 -> ConsumeEffect.Type.APPLY_EFFECTS
+        Registries.CONSUME_EFFECT_TYPE, registry -> ConsumeEffect.Type.APPLY_EFFECTS
     );
     public static final Registry<RecipeDisplay.Type<?>> RECIPE_DISPLAY = registerSimple(Registries.RECIPE_DISPLAY, RecipeDisplays::bootstrap);
     public static final Registry<SlotDisplay.Type<?>> SLOT_DISPLAY = registerSimple(Registries.SLOT_DISPLAY, SlotDisplays::bootstrap);
     public static final Registry<RecipeBookCategory> RECIPE_BOOK_CATEGORY = registerSimple(Registries.RECIPE_BOOK_CATEGORY, RecipeBookCategories::bootstrap);
-    public static final Registry<TicketType> TICKET_TYPE = registerSimple(Registries.TICKET_TYPE, var0 -> TicketType.UNKNOWN);
+    public static final Registry<TicketType> TICKET_TYPE = registerSimple(Registries.TICKET_TYPE, registry -> TicketType.UNKNOWN);
     public static final Registry<IncomingRpcMethod<?, ?>> INCOMING_RPC_METHOD = registerSimple(Registries.INCOMING_RPC_METHOD, IncomingRpcMethods::bootstrap);
     public static final Registry<OutgoingRpcMethod<?, ?>> OUTGOING_RPC_METHOD = registerSimple(
-        Registries.OUTGOING_RPC_METHOD, var0 -> OutgoingRpcMethods.SERVER_STARTED
+        Registries.OUTGOING_RPC_METHOD, registry -> OutgoingRpcMethods.SERVER_STARTED
     );
     public static final Registry<MapCodec<? extends TestEnvironmentDefinition<?>>> TEST_ENVIRONMENT_DEFINITION_TYPE = registerSimple(
         Registries.TEST_ENVIRONMENT_DEFINITION_TYPE, TestEnvironmentDefinition::bootstrap
@@ -364,9 +_,13 @@
     );
     public static final Registry<AttributeType<?>> ATTRIBUTE_TYPE = registerSimple(Registries.ATTRIBUTE_TYPE, AttributeTypes::bootstrap);
     public static final Registry<MapCodec<? extends SlotSource>> SLOT_SOURCE_TYPE = registerSimple(Registries.SLOT_SOURCE_TYPE, SlotSources::bootstrap);
-    public static final Registry<ContextKeySet> CONTEXT_KEY_SET = registerSimple(Registries.CONTEXT_KEY_SET, LootContextParamSets::bootstrap);
     public static final Registry<Consumer<GameTestHelper>> TEST_FUNCTION = registerSimple(Registries.TEST_FUNCTION, BuiltinTestFunctions::bootstrap);
     public static final Registry<? extends Registry<?>> REGISTRY = WRITABLE_REGISTRY;
+    // Paper start - add built-in registry conversions
+    public static final io.papermc.paper.registry.data.util.Conversions STATIC_ACCESS_CONVERSIONS = new io.papermc.paper.registry.data.util.Conversions(
+            new net.minecraft.resources.RegistryOps.HolderLookupAdapter(net.minecraft.server.RegistryLayer.STATIC_ACCESS)
+    );
+    // Paper end - add built-in registry conversions
 
     private static <T> Registry<T> registerSimple(ResourceKey<? extends Registry<T>> name, BuiltInRegistries.RegistryBootstrap<T> loader) {
         return internalRegister(name, new MappedRegistry<>(name, Lifecycle.stable(), false), loader);
@@ -392,6 +_,7 @@
         ResourceKey<? extends Registry<T>> name, R registry, BuiltInRegistries.RegistryBootstrap<T> loader
     ) {
         Bootstrap.checkBootstrapCalled(() -> "registry " + name.identifier());
+        io.papermc.paper.registry.PaperRegistryAccess.instance().registerRegistry(registry); // Paper - initialize API registry
         Identifier key = name.identifier();
         LOADERS.put(key, () -> loader.run(registry));
         WRITABLE_REGISTRY.register((ResourceKey)name, registry, RegistrationInfo.BUILT_IN);
@@ -399,16 +_,34 @@
     }
 
     public static void bootStrap() {
+        // Paper start
+        bootStrap(() -> {});
+    }
+    public static void bootStrap(Runnable runnable) {
+        // Paper end
+        REGISTRY.freeze(); // Paper - freeze main registry early
         createContents();
+        runnable.run(); // Paper
         freeze();
         validate(REGISTRY);
     }
 
     private static void createContents() {
+        // Paper start - class-load org.bukkit.Registry
+        // we have to class-load Registry here to create all the CraftRegistry instances
+        // that will be created when Registry is class-loaded before RegistryAccess#getRegistry
+        // would try to create them in lockReferenceHolder
+        try {
+            Class.forName(org.bukkit.Registry.class.getName());
+        } catch (final ClassNotFoundException ex) {
+            throw new RuntimeException(ex);
+        }
+        // Paper end - class-load org.bukkit.Registry
         LOADERS.forEach((key, value) -> {
             if (value.get() == null) {
                 LOGGER.error("Unable to bootstrap registry '{}'", key);
             }
+            io.papermc.paper.registry.PaperRegistryAccess.instance().lockReferenceHolders(ResourceKey.createRegistryKey(key)); // Paper - lock reference holder creation
         });
     }
 
@@ -417,6 +_,7 @@
 
         for (Registry<?> registry : REGISTRY) {
             bindBootstrappedTagsToEmpty(registry);
+            io.papermc.paper.registry.PaperRegistryListenerManager.INSTANCE.runFreezeListeners(registry.key(), STATIC_ACCESS_CONVERSIONS); // Paper
             registry.freeze();
         }
     }
@@ -440,6 +_,10 @@
 
     private static void bindBootstrappedTagsToEmpty(Registry<?> registry) {
         ((MappedRegistry)registry).bindAllTagsToEmpty();
+    }
+
+    public static java.util.Set<Identifier> getVanillaRegistrationOrder() {
+        return java.util.Collections.unmodifiableSet(LOADERS.keySet());
     }
 
     @FunctionalInterface
