--- a/net/minecraft/data/structures/StructureUpdater.java
+++ b/net/minecraft/data/structures/StructureUpdater.java
@@ -22,12 +_,12 @@
     public static CompoundTag update(String name, CompoundTag tag) {
         StructureTemplate structureTemplate = new StructureTemplate();
         int fromVersion = NbtUtils.getDataVersion(tag, 500);
-        int toVersion = 4997;
-        if (fromVersion < 4997) {
-            LOGGER.warn("SNBT Too old, do not forget to update: {} < {}: {}", fromVersion, 4997, name);
+        int toVersion = 4882;
+        if (fromVersion < 4882) {
+            LOGGER.warn("SNBT Too old, do not forget to update: {} < {}: {}", fromVersion, 4882, name);
         }
 
-        CompoundTag updated = DataFixTypes.STRUCTURE.updateToCurrentVersion(DataFixers.getDataFixer(), tag, fromVersion);
+        CompoundTag updated = ca.spottedleaf.dataconverter.minecraft.MCDataConverter.convertTag(ca.spottedleaf.dataconverter.minecraft.datatypes.MCTypeRegistry.STRUCTURE, tag, fromVersion, ca.spottedleaf.dataconverter.minecraft.util.Version.getCurrentVersion()); // Paper - rewrite dataconverter system
         structureTemplate.load(BuiltInRegistries.BLOCK, updated);
         return structureTemplate.save(new CompoundTag());
     }
