--- a/net/minecraft/world/entity/animal/equine/AbstractHorse.java
+++ b/net/minecraft/world/entity/animal/equine/AbstractHorse.java
@@ -27,7 +_,6 @@
 import net.minecraft.world.entity.AgeableMob;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityDimensions;
-import net.minecraft.world.entity.EntityEvent;
 import net.minecraft.world.entity.EntityReference;
 import net.minecraft.world.entity.EntitySpawnReason;
 import net.minecraft.world.entity.EntityType;
@@ -75,6 +_,10 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRegainHealthEvent;
+// CraftBukkit end
 import org.jspecify.annotations.Nullable;
 
 public abstract class AbstractHorse extends Animal implements PlayerRideableJumping, HasCustomInventoryScreen, OwnableEntity {
@@ -124,15 +_,65 @@
     protected boolean canGallop = true;
     protected int gallopSoundCounter;
     public @Nullable EntityReference<LivingEntity> owner;
+    public int maxDomestication = 100; // CraftBukkit - store max domestication value
 
     protected AbstractHorse(EntityType<? extends AbstractHorse> type, Level level) {
         super(type, level);
+        this.moveControl = new net.minecraft.world.entity.ai.control.MoveControl(this); // Purpur - use vanilla controller
+        this.lookControl = new net.minecraft.world.entity.ai.control.LookControl(this); // Purpur - use vanilla controller
         this.createInventory();
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return false; // vanilla handles
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.generateMaxHealth(random));
+        this.getAttribute(Attributes.MOVEMENT_SPEED).setBaseValue(this.generateSpeed(random));
+        this.getAttribute(Attributes.JUMP_STRENGTH).setBaseValue(this.generateJumpStrength(random));
+    }
+
+    protected double generateMaxHealth(double min, double max) {
+        if (min == max) return min;
+        int diff = Mth.floor(max - min);
+        double base = max - diff;
+        int first = Mth.floor((double) diff / 2);
+        int rest = diff - first;
+        return base + random.nextInt(first + 1) + random.nextInt(rest + 1);
+    }
+
+    protected double generateJumpStrength(double min, double max) {
+        if (min == max) return min;
+        return min + (max - min) * this.random.nextDouble();
+    }
+
+    protected double generateSpeed(double min, double max) {
+        if (min == max) return min;
+        return min + (max - min) * this.random.nextDouble();
+    }
+
+    protected float generateMaxHealth(RandomSource random) {
+        return 15.0F + (float) random.nextInt(8) + (float) random.nextInt(9);
+    }
+
+    protected double generateJumpStrength(RandomSource random) {
+        return 0.4F + random.nextDouble() * 0.2 + random.nextDouble() * 0.2 + random.nextDouble() * 0.2;
+    }
+
+    protected double generateSpeed(RandomSource random) {
+        return (0.45F + random.nextDouble() * 0.3 + random.nextDouble() * 0.3 + random.nextDouble() * 0.3) * 0.25;
+    }
+    // Purpur end - Configurable entity base attributes
+
     @Override
     protected void registerGoals() {
-        this.goalSelector.addGoal(1, new RunAroundLikeCrazyGoal(this, 1.2));
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HorseHasRider(this)); // Purpur - Ridables
         this.goalSelector.addGoal(2, new BreedGoal(this, 1.0, AbstractHorse.class));
         this.goalSelector.addGoal(4, new FollowParentGoal(this, 1.0));
         this.goalSelector.addGoal(6, new WaterAvoidingRandomStrollGoal(this, 0.7));
@@ -252,7 +_,7 @@
     }
 
     @Override
-    public boolean isPushable() {
+    public boolean isCollidable(boolean ignoreClimbing) { // Paper - Climbing should not bypass cramming gamerule
         return !this.isVehicle();
     }
 
@@ -299,7 +_,7 @@
 
     public void createInventory() {
         SimpleContainer old = this.inventory;
-        this.inventory = new SimpleContainer(this.getInventorySize());
+        this.inventory = new SimpleContainer(this.getInventorySize(), (org.bukkit.entity.AbstractHorse) this.getBukkitEntity()); // CraftBukkit
         if (old != null) {
             int max = Math.min(old.getContainerSize(), this.inventory.getContainerSize());
 
@@ -343,9 +_,9 @@
     protected void playStepSound(BlockPos pos, BlockState blockState) {
         if (!blockState.liquid()) {
             BlockState aboveState = this.level().getBlockState(pos.above());
-            SoundType soundType = blockState.getSoundType();
+            SoundType soundType = blockState.getSoundType(level(), pos, this);
             if (aboveState.is(Blocks.SNOW)) {
-                soundType = aboveState.getSoundType();
+                soundType = aboveState.getSoundType(level(), pos, this);
             }
 
             if (this.isVehicle() && this.canGallop) {
@@ -391,7 +_,7 @@
     }
 
     public int getMaxTemper() {
-        return 100;
+        return this.maxDomestication; // CraftBukkit - return stored max domestication instead of 100
     }
 
     @Override
@@ -454,7 +_,7 @@
             temper = 5;
             if (!this.level().isClientSide() && this.isTamed() && this.getAge() == 0 && !this.isInLove()) {
                 itemUsed = true;
-                this.setInLove(player);
+                this.setInLove(player, itemStack.copy()); // Paper - Fix EntityBreedEvent copying
             }
         } else if (itemStack.is(Items.GOLDEN_APPLE) || itemStack.is(Items.ENCHANTED_GOLDEN_APPLE)) {
             heal = 10.0F;
@@ -462,12 +_,12 @@
             temper = 10;
             if (!this.level().isClientSide() && this.isTamed() && this.getAge() == 0 && !this.isInLove()) {
                 itemUsed = true;
-                this.setInLove(player);
+                this.setInLove(player, itemStack.copy()); // Paper - Fix EntityBreedEvent copying
             }
         }
 
         if (this.getHealth() < this.getMaxHealth() && heal > 0.0F) {
-            this.heal(heal);
+            this.heal(heal, EntityRegainHealthEvent.RegainReason.EATING); // CraftBukkit
             itemUsed = true;
         }
 
@@ -538,7 +_,7 @@
         super.aiStep();
         if (this.level() instanceof ServerLevel level && this.isAlive()) {
             if (this.random.nextInt(900) == 0 && this.deathTime == 0) {
-                this.heal(1.0F);
+                this.heal(1.0F, EntityRegainHealthEvent.RegainReason.REGEN); // CraftBukkit
             }
 
             if (this.canEatGrass()) {
@@ -640,6 +_,16 @@
         }
     }
 
+    // Paper start - Horse API
+    public void setMouthOpen(boolean open) {
+        this.setFlag(FLAG_OPEN_MOUTH, open);
+    }
+
+    public boolean isMouthOpen() {
+        return this.getFlag(FLAG_OPEN_MOUTH);
+    }
+    // Paper end - Horse API
+
     @Override
     public InteractionResult mobInteract(Player player, InteractionHand hand) {
         if (this.isVehicle() || this.isBaby()) {
@@ -774,6 +_,7 @@
         Vec3 movement = this.getDeltaMovement();
         this.setDeltaMovement(movement.x, impulse, movement.z);
         this.needsSync = true;
+        net.neoforged.neoforge.common.CommonHooks.onLivingJump(this);
         if (input.z > 0.0) {
             float sin = Mth.sin(this.getYRot() * (float) (Math.PI / 180.0));
             float cos = Mth.cos(this.getYRot() * (float) (Math.PI / 180.0));
@@ -793,6 +_,7 @@
         output.putInt("Temper", this.getTemper());
         output.putBoolean("Tame", this.isTamed());
         EntityReference.store(this.owner, output, "Owner");
+        output.putInt("Bukkit.MaxDomestication", this.maxDomestication); // CraftBukkit
     }
 
     @Override
@@ -803,6 +_,9 @@
         this.setTemper(input.getIntOr("Temper", 0));
         this.setTamed(input.getBooleanOr("Tame", false));
         this.owner = EntityReference.readWithOldOwnerConversion(input, "Owner", this.level());
+        // CraftBukkit start
+        this.maxDomestication = input.getIntOr("Bukkit.MaxDomestication", this.maxDomestication);
+        // CraftBukkit end
     }
 
     @Override
@@ -894,6 +_,17 @@
 
     @Override
     public void handleStartJump(int jumpScale) {
+        // CraftBukkit start
+        float power;
+        if (jumpScale >= 90) {
+            power = 1.0F;
+        } else {
+            power = 0.4F + 0.4F * (float) jumpScale / 90.0F;
+        }
+        if (!CraftEventFactory.callHorseJumpEvent(this, power)) {
+            return;
+        }
+        // CraftBukkit end
         this.allowStandSliding = true;
         this.standIfPossible();
         this.playJumpSound();
@@ -915,7 +_,7 @@
     }
 
     @Override
-    public void handleEntityEvent(@EntityEvent.Value byte id) {
+    public void handleEntityEvent(byte id) {
         if (id == 7) {
             this.spawnTamingParticles(true);
         } else if (id == 6) {
@@ -1021,8 +_,13 @@
             groupData = new AgeableMob.AgeableMobGroupData(0.2F);
         }
 
-        this.randomizeAttributes(level.getRandom());
+        //this.randomizeAttributes(level.getRandom()); // Purpur - replaced by initAttributes()
         return super.finalizeSpawn(level, difficulty, spawnReason, groupData);
+    }
+
+    // Neo: Inventory getter
+    public net.minecraft.world.Container getInventory() {
+        return this.inventory;
     }
 
     public boolean hasInventoryChanged(Container oldInventory) {
