--- a/net/minecraft/world/entity/animal/equine/Llama.java
+++ b/net/minecraft/world/entity/animal/equine/Llama.java
@@ -75,7 +_,8 @@
         .scale(0.5F);
     private boolean didSpit;
     private @Nullable Llama caravanHead;
-    private @Nullable Llama caravanTail;
+    @Nullable
+    public Llama caravanTail;
 
     public Llama(EntityType<? extends Llama> type, Level level) {
         super(type, level);
@@ -86,7 +_,13 @@
         return false;
     }
 
-    private void setStrength(int strength) {
+    // CraftBukkit start
+    public void setStrengthPublic(int i) {
+        this.setStrength(i);
+    }
+    // CraftBukkit end
+
+    public void setStrength(int strength) {
         this.entityData.set(DATA_STRENGTH_ID, Math.max(1, Math.min(5, strength)));
     }
 
