--- a/net/minecraft/world/entity/animal/equine/Llama.java
+++ b/net/minecraft/world/entity/animal/equine/Llama.java
@@ -86,6 +_,12 @@
         return false;
     }
 
+    // CraftBukkit start
+    public void setStrengthPublic(int i) {
+        this.setStrength(i);
+    }
+    // CraftBukkit end
+
     private void setStrength(int strength) {
         this.entityData.set(DATA_STRENGTH_ID, Math.max(1, Math.min(5, strength)));
     }
