--- a/net/minecraft/world/entity/animal/equine/ZombieHorse.java
+++ b/net/minecraft/world/entity/animal/equine/ZombieHorse.java
@@ -48,6 +_,7 @@
 
     public ZombieHorse(EntityType<? extends ZombieHorse> type, Level level) {
         super(type, level);
+        this.setShouldBurnInDay(true); // Purpur - API for any mob to burn daylight
     }
 
     public static AttributeSupplier.Builder createAttributes() {
