--- a/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacer.java
@@ -169,7 +_,7 @@
     protected static boolean tryPlaceLeaf(WorldGenLevel level, FoliagePlacer.FoliageSetter foliageSetter, RandomSource random, TreeFeature tree, BlockPos pos) {
         boolean isPersistent = level.isStateAtPosition(pos, state -> state.getValueOrElse(BlockStateProperties.PERSISTENT, false));
         if (!isPersistent && TreeFeature.validTreePos(level, pos)) {
-            BlockState foliageState = tree.foliageProvider().value().getState(level, random, pos);
+            BlockState foliageState = tree.foliageProvider().getState(level, random, pos);
             if (foliageState.hasProperty(BlockStateProperties.WATERLOGGED)) {
                 foliageState = foliageState.setValue(
                     BlockStateProperties.WATERLOGGED, level.isFluidAtPosition(pos, fluidState -> fluidState.isSourceOfType(Fluids.WATER))
