--- a/net/minecraft/world/level/levelgen/feature/foliageplacers/PoplarFoliagePlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/foliageplacers/PoplarFoliagePlacer.java
@@ -158,8 +_,8 @@
         BlockPos pos,
         Function<BlockState, BlockState> stateModifier
     ) {
-        if (level.isStateAtPosition(pos, state -> state.equals(tree.foliageProvider().value().getState(level, random, pos)))) {
-            foliageSetter.set(pos, stateModifier.apply(tree.trunkProvider().value().getState(level, random, pos)));
+        if (level.isStateAtPosition(pos, state -> state.equals(tree.foliageProvider().getState(level, random, pos)))) {
+            foliageSetter.set(pos, stateModifier.apply(tree.trunkProvider().getState(level, random, pos)));
         }
     }
 
