--- a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
+++ b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
@@ -25,9 +_,7 @@
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.InterpolationHandler;
 import net.minecraft.world.entity.Leashable;
-import net.minecraft.world.entity.LinearInterpolationHandler;
 import net.minecraft.world.entity.LivingEntity;
-import net.minecraft.world.entity.Mob;
 import net.minecraft.world.entity.MoverType;
 import net.minecraft.world.entity.Pose;
 import net.minecraft.world.entity.animal.Animal;
@@ -65,7 +_,7 @@
     private final float[] paddlePositions = new float[2];
     private float outOfControlTicks;
     private float deltaRotation;
-    private final InterpolationHandler interpolation = new LinearInterpolationHandler(this, 3);
+    private final InterpolationHandler interpolation = new InterpolationHandler(this, 3);
     private boolean inputLeft;
     private boolean inputRight;
     private boolean inputUp;
@@ -83,6 +_,15 @@
     private Leashable.@Nullable LeashData leashData;
     private final Supplier<Item> dropItem;
 
+    // CraftBukkit start
+    // PAIL: Some of these haven't worked since a few updates, and since 1.9 they are less and less applicable.
+    public double maxSpeed = 0.4D;
+    public double occupiedDeceleration = 0.2D;
+    public double unoccupiedDeceleration = -1;
+    public boolean landBoats = false;
+    private org.bukkit.Location lastLocation;
+    // CraftBukkit end
+
     public AbstractBoat(EntityType<? extends AbstractBoat> type, Level level, Supplier<Item> dropItem) {
         super(type, level);
         this.dropItem = dropItem;
@@ -124,7 +_,7 @@
     }
 
     @Override
-    public boolean isPushable() {
+    public boolean isCollidable(boolean ignoreClimbing) { // Paper - Climbing should not bypass cramming gamerule
         return true;
     }
 
@@ -179,11 +_,30 @@
 
     @Override
     public void push(Entity entity) {
+        if (!this.level().paperConfig().collisions.allowVehicleCollisions && this.level().paperConfig().collisions.onlyPlayersCollide && !(entity instanceof Player)) return; // Paper - Collision option for requiring a player participant
         if (entity instanceof AbstractBoat) {
             if (entity.getBoundingBox().minY < this.getBoundingBox().maxY) {
+                // CraftBukkit start
+                if (!this.isPassengerOfSameVehicle(entity)) {
+                    org.bukkit.event.vehicle.VehicleEntityCollisionEvent event = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                        (org.bukkit.entity.Vehicle) this.getBukkitEntity(),
+                        entity.getBukkitEntity()
+                    );
+                    if (!event.callEvent()) return;
+                }
+                // CraftBukkit end
                 super.push(entity);
             }
         } else if (entity.getBoundingBox().minY <= this.getBoundingBox().minY) {
+            // CraftBukkit start
+            if (!this.isPassengerOfSameVehicle(entity)) {
+                org.bukkit.event.vehicle.VehicleEntityCollisionEvent event = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
+                    (org.bukkit.entity.Vehicle) this.getBukkitEntity(),
+                    entity.getBukkitEntity()
+                );
+                if (!event.callEvent()) return;
+            }
+            // CraftBukkit end
             super.push(entity);
         }
     }
@@ -250,6 +_,18 @@
             this.setDeltaMovement(Vec3.ZERO);
         }
 
+        // CraftBukkit start
+        org.bukkit.Location to = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(this.position(), this.level(), this.getYRot(), this.getXRot());
+        org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
+
+        new org.bukkit.event.vehicle.VehicleUpdateEvent(vehicle).callEvent();
+
+        if (this.lastLocation != null && !this.lastLocation.equals(to)) {
+            org.bukkit.event.vehicle.VehicleMoveEvent event = new org.bukkit.event.vehicle.VehicleMoveEvent(vehicle, this.lastLocation, to);
+            event.callEvent();
+        }
+        this.lastLocation = vehicle.getLocation();
+        // CraftBukkit end
         this.applyEffectsFromBlocks();
         this.applyEffectsFromBlocks();
         this.tickBubbleColumn();
@@ -328,14 +_,13 @@
                 int diff = 60 - bubbleTime - 1;
                 if (diff > 0 && bubbleTime == 0) {
                     this.setBubbleTime(0);
+                    Vec3 movement = this.getDeltaMovement();
                     if (this.bubbleColumnDirectionIsDown) {
+                        this.setDeltaMovement(movement.add(0.0, -0.7, 0.0));
                         this.ejectPassengers();
-                        this.level().broadcastEntityEvent(this, (byte)71);
                     } else {
-                        this.level().broadcastEntityEvent(this, (byte)72);
+                        this.setDeltaMovement(movement.x, this.hasPassenger(e -> e instanceof Player) ? 2.7 : 0.6, movement.z);
                     }
-
-                    this.handleBubbleColumnEffect(this.bubbleColumnDirectionIsDown);
                 }
 
                 this.isAboveBubbleColumn = false;
@@ -343,27 +_,6 @@
         }
     }
 
-    @Override
-    public void handleEntityEvent(byte id) {
-        super.handleEntityEvent(id);
-        if (id == 71) {
-            this.handleBubbleColumnEffect(true);
-        } else if (id == 72) {
-            this.handleBubbleColumnEffect(false);
-        }
-    }
-
-    protected void handleBubbleColumnEffect(boolean dragDown) {
-        if (this.canSimulateMovement()) {
-            Vec3 movement = this.getDeltaMovement();
-            if (dragDown) {
-                this.setDeltaMovement(movement.add(0.0, -0.7, 0.0));
-            } else {
-                this.setDeltaMovement(movement.x, 0.6, movement.z);
-            }
-        }
-    }
-
     protected @Nullable SoundEvent getPaddleSound() {
         return switch (this.getStatus()) {
             case IN_WATER, UNDER_WATER, UNDER_FLOWING_WATER -> SoundEvents.BOAT_PADDLE_WATER;
@@ -417,6 +_,7 @@
             float friction = this.getGroundFriction();
             if (friction > 0.0F) {
                 this.landFriction = friction;
+                if (level().purpurConfig.boatEjectPlayersOnLand) ejectPassengers(); // Purpur - Add option for boats to eject players on land
                 return AbstractBoat.Status.ON_LAND;
             } else {
                 return AbstractBoat.Status.IN_AIR;
@@ -442,7 +_,7 @@
                 for (int z = minZ; z < maxZ; z++) {
                     pos.set(x, y, z);
                     FluidState fluidState = this.level().getFluidState(pos);
-                    if (fluidState.is(FluidTags.WATER)) {
+                    if (this.canBoatInFluid(fluidState)) {
                         blockHeight = Math.max(blockHeight, fluidState.getHeight(this.level(), pos));
                     }
 
@@ -484,7 +_,7 @@
                             BlockState blockState = this.level().getBlockState(blockPos);
                             if (!(blockState.getBlock() instanceof LilyPadBlock)
                                 && Shapes.joinIsNotEmpty(blockState.getCollisionShape(this.level(), blockPos).move(blockPos), boatShape, BooleanOp.AND)) {
-                                friction += blockState.getBlock().getFriction();
+                                friction += blockState.getFriction(this.level(), blockPos, this);
                                 count++;
                             }
                         }
@@ -513,7 +_,7 @@
                 for (int z = minZ; z < maxZ; z++) {
                     pos.set(x, y, z);
                     FluidState fluidState = this.level().getFluidState(pos);
-                    if (fluidState.is(FluidTags.WATER)) {
+                    if (this.canBoatInFluid(fluidState)) {
                         float height = y + fluidState.getHeight(this.level(), pos);
                         this.waterLevel = Math.max(height, this.waterLevel);
                         inWater |= bb.minY < height;
@@ -542,7 +_,7 @@
                 for (int z = z0; z < z1; z++) {
                     pos.set(x, y, z);
                     FluidState fluidState = this.level().getFluidState(pos);
-                    if (fluidState.is(FluidTags.WATER) && maxY < pos.getY() + fluidState.getHeight(this.level(), pos)) {
+                    if (this.canBoatInFluid(fluidState) && maxY < pos.getY() + fluidState.getHeight(this.level(), pos)) {
                         if (!fluidState.isSource()) {
                             return AbstractBoat.Status.UNDER_FLOWING_WATER;
                         }
@@ -569,7 +_,7 @@
             this.waterLevel = this.getY(1.0);
             double targetY = this.getWaterLevelAbove() - this.getBbHeight() + 0.101;
             if (this.level().noCollision(this, this.getBoundingBox().move(0.0, targetY - this.getY(), 0.0))) {
-                this.setPos(this.getX(), targetY, this.getZ());
+                this.move(MoverType.SELF, new Vec3(0.0, targetY - this.getY(), 0.0)); // Paper - Fix some exploit with boats // TODO Still needed?
                 this.setDeltaMovement(this.getDeltaMovement().multiply(1.0, 0.0, 1.0));
                 this.lastYd = 0.0;
             }
@@ -660,6 +_,11 @@
             passenger.setYRot(passenger.getYRot() + this.deltaRotation);
             passenger.setYHeadRot(passenger.getYHeadRot() + this.deltaRotation);
             this.clampRotation(passenger);
+            if (passenger instanceof Animal animal && this.getPassengers().size() == this.getMaxPassengers()) {
+                int rotationOffset = passenger.getId() % 2 == 0 ? 90 : 270;
+                passenger.setYBodyRot(animal.yBodyRot + rotationOffset);
+                passenger.setYHeadRot(passenger.getYHeadRot() + rotationOffset);
+            }
         }
     }
 
@@ -696,26 +_,12 @@
     }
 
     protected void clampRotation(Entity passenger) {
-        float passengerBodyYRot = this.calculatePassengerBodyYRot(passenger);
-        passenger.setYBodyRot(passengerBodyYRot);
-        if (passenger instanceof Mob mob) {
-            passenger.yRotO = passenger.yRotO + Mth.wrapDegrees(passengerBodyYRot - passenger.getYRot());
-            passenger.setYRot(passengerBodyYRot);
-            mob.clampHeadRotationToBody();
-        } else {
-            float delta = Mth.wrapDegrees(passenger.getYRot() - passengerBodyYRot);
-            float maxHeadRotation = 105.0F;
-            float targetDelta = Mth.clamp(delta, -105.0F, 105.0F);
-            passenger.yRotO += targetDelta - delta;
-            passenger.setYRot(passenger.getYRot() + targetDelta - delta);
-            passenger.setYHeadRot(passenger.getYRot());
-        }
-    }
-
-    private float calculatePassengerBodyYRot(Entity passenger) {
-        return passenger instanceof Animal && this.getPassengers().size() == this.getMaxPassengers()
-            ? this.getYRot() + (passenger.getId() % 2 == 0 ? 90.0F : 270.0F)
-            : this.getYRot();
+        passenger.setYBodyRot(this.getYRot());
+        float delta = Mth.wrapDegrees(passenger.getYRot() - this.getYRot());
+        float targetDelta = Mth.clamp(delta, -105.0F, 105.0F);
+        passenger.yRotO += targetDelta - delta;
+        passenger.setYRot(passenger.getYRot() + targetDelta - delta);
+        passenger.setYHeadRot(passenger.getYRot());
     }
 
     @Override
@@ -746,12 +_,20 @@
     }
 
     @Override
-    public void remove(Entity.RemovalReason reason) {
+    public void remove(final Entity.RemovalReason reason, org.bukkit.event.entity.EntityRemoveEvent.@Nullable Cause eventCause) { // CraftBukkit - add Bukkit remove cause
         if (!this.level().isClientSide() && reason.shouldDestroy() && this.isLeashed()) {
-            this.dropLeash();
+            // Paper start - Fire EntityUnleashEvent on boat destruction
+            final org.bukkit.event.entity.EntityUnleashEvent event = new org.bukkit.event.entity.EntityUnleashEvent(this.getBukkitEntity(), org.bukkit.event.entity.EntityUnleashEvent.UnleashReason.LEASHED_GONE, true);
+            event.callEvent();
+            if (event.isDropLeash()) {
+                this.dropLeash();
+            } else {
+                this.removeLeash();
+            }
+            // Paper end - Fire EntityUnleashEvent on boat destruction
         }
 
-        super.remove(reason);
+        super.remove(reason, eventCause); // CraftBukkit - add Bukkit remove cause
     }
 
     @Override
@@ -760,7 +_,7 @@
         if (!this.isPassenger()) {
             if (onGround) {
                 this.resetFallDistance();
-            } else if (!this.level().getFluidState(this.blockPosition().below()).is(FluidTags.WATER) && ya < 0.0) {
+            } else if (!this.canBoatInFluid(this.level().getFluidState(this.blockPosition().below())) && ya < 0.0) {
                 this.fallDistance -= (float)ya;
             }
         }
@@ -815,7 +_,13 @@
 
     @Override
     public final ItemStack getPickResult() {
-        return new ItemStack(this.dropItem.get());
+        // Purpur start - Apply display names from item forms of entities to entities and vice versa
+        final ItemStack boat = new ItemStack(this.dropItem.get());
+        if (!this.level().purpurConfig.persistentDroppableEntityDisplayNames) {
+            boat.set(net.minecraft.core.component.DataComponents.CUSTOM_NAME, null);
+        }
+        return boat;
+        // Purpur end - Apply display names from item forms of entities to entities and vice versa
     }
 
     @Override
