--- a/net/minecraft/world/level/levelgen/structure/pieces/StructurePieceSerializationContext.java
+++ b/net/minecraft/world/level/levelgen/structure/pieces/StructurePieceSerializationContext.java
@@ -11,6 +_,6 @@
 ) {
     public static StructurePieceSerializationContext fromLevel(ServerLevel level) {
         MinecraftServer server = level.getServer();
-        return new StructurePieceSerializationContext(server.getResourceManager(), server.registryAccess(), server.getStructureTemplateManager());
+        return new StructurePieceSerializationContext(server.getResourceManager(), server.registryAccess(), server.getStructureManager());
     }
 }
