--- a/net/minecraft/world/level/levelgen/structure/pieces/StructurePiecesBuilder.java
+++ b/net/minecraft/world/level/levelgen/structure/pieces/StructurePiecesBuilder.java
@@ -5,15 +_,18 @@
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
 import net.minecraft.world.level.levelgen.structure.StructurePiece;
+import net.minecraft.world.level.levelgen.structure.StructurePieceAccessor;
 import org.jspecify.annotations.Nullable;
 
-public class StructurePiecesBuilder {
+public class StructurePiecesBuilder implements StructurePieceAccessor {
     private final List<StructurePiece> pieces = Lists.newArrayList();
 
+    @Override
     public void addPiece(StructurePiece piece) {
         this.pieces.add(piece);
     }
 
+    @Override
     public @Nullable StructurePiece findCollisionPiece(BoundingBox box) {
         return StructurePiece.findCollisionPiece(this.pieces, box);
     }
