--- /dev/null
+++ b/net/minecraft/world/level/storage/loot/IntRange.java
@@ -1,0 +_,114 @@
+package net.minecraft.world.level.storage.loot;
+
+import com.mojang.datafixers.util.Either;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.function.Function;
+import net.minecraft.core.Holder;
+import net.minecraft.util.Mth;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
+import org.jspecify.annotations.Nullable;
+
+public class IntRange implements LootContextUser {
+    private static final Codec<IntRange> RECORD_CODEC = RecordCodecBuilder.create(
+        i -> i.group(
+                NumberProviders.CODEC.optionalFieldOf("min").forGetter(r -> Optional.ofNullable(r.min)),
+                NumberProviders.CODEC.optionalFieldOf("max").forGetter(r -> Optional.ofNullable(r.max))
+            )
+            .apply(i, IntRange::new)
+    );
+    public static final Codec<IntRange> CODEC = Codec.either(Codec.INT, RECORD_CODEC).xmap(e -> e.map(IntRange::exact, Function.identity()), range -> {
+        OptionalInt exact = range.unpackExact();
+        return exact.isPresent() ? Either.left(exact.getAsInt()) : Either.right(range);
+    });
+    private final @Nullable Holder<NumberProvider> min;
+    private final @Nullable Holder<NumberProvider> max;
+    private final IntRange.IntLimiter limiter;
+    private final IntRange.IntChecker predicate;
+
+    @Override
+    public void validate(ValidationContext context) {
+        LootContextUser.super.validate(context);
+        if (this.min != null) {
+            Validatable.validateHolder(context, "min", this.min);
+        }
+
+        if (this.max != null) {
+            Validatable.validateHolder(context, "max", this.max);
+        }
+    }
+
+    private IntRange(Optional<Holder<NumberProvider>> min, Optional<Holder<NumberProvider>> max) {
+        this(min.orElse(null), max.orElse(null));
+    }
+
+    private IntRange(@Nullable Holder<NumberProvider> min, @Nullable Holder<NumberProvider> max) {
+        this.min = min;
+        this.max = max;
+        if (min == null) {
+            if (max == null) {
+                this.limiter = (context, value) -> value;
+                this.predicate = (context, value) -> true;
+            } else {
+                this.limiter = (context, value) -> Math.min(max.value().getInt(context), value);
+                this.predicate = (context, value) -> value <= max.value().getInt(context);
+            }
+        } else if (max == null) {
+            this.limiter = (context, value) -> Math.max(min.value().getInt(context), value);
+            this.predicate = (context, value) -> value >= min.value().getInt(context);
+        } else {
+            this.limiter = (context, value) -> Mth.clamp(value, min.value().getInt(context), max.value().getInt(context));
+            this.predicate = (context, value) -> value >= min.value().getInt(context) && value <= max.value().getInt(context);
+        }
+    }
+
+    public static IntRange exact(int value) {
+        Holder<NumberProvider> c = ConstantValue.exactly(value);
+        return new IntRange(Optional.of(c), Optional.of(c));
+    }
+
+    public static IntRange range(int min, int max) {
+        return new IntRange(Optional.of(ConstantValue.exactly(min)), Optional.of(ConstantValue.exactly(max)));
+    }
+
+    public static IntRange lowerBound(int value) {
+        return new IntRange(Optional.of(ConstantValue.exactly(value)), Optional.empty());
+    }
+
+    public static IntRange upperBound(int value) {
+        return new IntRange(Optional.empty(), Optional.of(ConstantValue.exactly(value)));
+    }
+
+    public int clamp(LootContext context, int value) {
+        return this.limiter.apply(context, value);
+    }
+
+    public boolean test(LootContext context, int value) {
+        return this.predicate.test(context, value);
+    }
+
+    private OptionalInt unpackExact() {
+        return Objects.equals(this.min, this.max)
+                && this.min != null
+                && this.min instanceof Holder.Direct
+                && this.min.value() instanceof ConstantValue constant
+                && Math.floor(constant.value()) == constant.value()
+            ? OptionalInt.of((int)constant.value())
+            : OptionalInt.empty();
+    }
+
+    @FunctionalInterface
+    private interface IntChecker {
+        boolean test(LootContext context, int value);
+    }
+
+    @FunctionalInterface
+    private interface IntLimiter {
+        int apply(LootContext context, int value);
+    }
+}
