--- a/net/minecraft/world/level/storage/loot/ValidationContext.java
+++ b/net/minecraft/world/level/storage/loot/ValidationContext.java
@@ -89,13 +_,6 @@
         return this.reporter;
     }
 
-    public record MinBoundsProblem(int minBounds) implements ProblemReporter.Problem {
-        @Override
-        public String description() {
-            return "List must contain at least " + this.minBounds + " element(s)";
-        }
-    }
-
     public record ParametersNotProvidedProblem(Set<ContextKey<?>> notProvided) implements ProblemReporter.Problem {
         @Override
         public String description() {
@@ -108,22 +_,12 @@
         public String description() {
             return this.referenced.identifier() + " of type " + this.referenced.registry() + " is recursively called";
         }
-
-        @Override
-        public boolean isFatal() {
-            return true;
-        }
     }
 
     public record RecursiveTagReferenceProblem(TagKey<?> referenced) implements ProblemReporter.Problem {
         @Override
         public String description() {
             return "#" + this.referenced.location() + " of type " + this.referenced.registry() + " is recursively called";
-        }
-
-        @Override
-        public boolean isFatal() {
-            return true;
         }
     }
 }
