--- a/net/minecraft/resources/Identifier.java
+++ b/net/minecraft/resources/Identifier.java
@@ -17,18 +_,26 @@
 
 public final class Identifier implements Comparable<Identifier> {
     public static final Codec<Identifier> CODEC = Codec.STRING.<Identifier>comapFlatMap(Identifier::read, Identifier::toString).stable();
-    public static final StreamCodec<ByteBuf, Identifier> STREAM_CODEC = ByteBufCodecs.STRING_UTF8.map(Identifier::parse, Identifier::toString);
+    public static final StreamCodec<ByteBuf, Identifier> STREAM_CODEC = ByteBufCodecs.STRING_UTF8.map(Identifier::parse, rl -> rl.getNamespace().equals("minecraft") ? rl.getPath() : rl.toString());
     public static final SimpleCommandExceptionType ERROR_INVALID = new SimpleCommandExceptionType(Component.translatable("argument.id.invalid"));
     public static final char NAMESPACE_SEPARATOR = ':';
     public static final String DEFAULT_NAMESPACE = "minecraft";
     public static final String REALMS_NAMESPACE = "realms";
     public static final String ALLOWED_NAMESPACE_CHARACTERS = "[a-z0-9_.-]";
+    public static final String PAPER_NAMESPACE = "paper"; // Paper
     private final String namespace;
     private final String path;
 
     private Identifier(String namespace, String path) {
         assert isValidNamespace(namespace);
         assert isValidPath(path);
+        // Paper start - Validate Identifier
+        // Check for the max network string length (capped at Short.MAX_VALUE) as well as the max bytes of a StringTag (length written as an unsigned short)
+        final int length = namespace.length() + path.length() + 1;
+        if (length > Short.MAX_VALUE || io.netty.buffer.ByteBufUtil.utf8MaxBytes(length) > 2 * Short.MAX_VALUE + 1) {
+            throw new IdentifierException("Identifier too long: " + length);
+        }
+        // Paper end - Validate Identifier
         this.namespace = namespace;
         this.path = path;
     }
@@ -127,7 +_,7 @@
     }
 
     @Override
-    public boolean equals(@Nullable Object o) {
+    public boolean equals(Object o) {
         if (this == o) {
             return true;
         } else {
@@ -140,6 +_,12 @@
         return 31 * this.namespace.hashCode() + this.path.hashCode();
     }
 
+    // Normal compare sorts by path first, this compares namespace first.
+    public int compareNamespaced(Identifier o) {
+        int ret = this.namespace.compareTo(o.namespace);
+        return ret != 0 ? ret : this.path.compareTo(o.path);
+    }
+
     public int compareTo(Identifier o) {
         int result = this.path.compareTo(o.path);
         if (result == 0) {
@@ -253,7 +_,7 @@
 
     private static String assertValidNamespace(String namespace, String path) {
         if (!isValidNamespace(namespace)) {
-            throw new IdentifierException("Non [a-z0-9_.-] character in namespace of identifier: " + namespace + ":" + path);
+            throw new IdentifierException("Non [a-z0-9_.-] character in namespace of identifier: " + org.apache.commons.lang3.StringUtils.normalizeSpace(namespace) + ":" + org.apache.commons.lang3.StringUtils.normalizeSpace(path)); // Paper - Sanitize Identifier error logging
         } else {
             return namespace;
         }
@@ -269,7 +_,7 @@
 
     private static String assertValidPath(String namespace, String path) {
         if (!isValidPath(path)) {
-            throw new IdentifierException("Non [a-z0-9/._-] character in path of location: " + namespace + ":" + path);
+            throw new IdentifierException("Non [a-z0-9/._-] character in path of location: " + namespace + ":" + org.apache.commons.lang3.StringUtils.normalizeSpace(path)); // Paper - Sanitize Identifier error logging
         } else {
             return path;
         }
