--- a/net/minecraft/resources/RegistryOps.java
+++ b/net/minecraft/resources/RegistryOps.java
@@ -11,6 +_,7 @@
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderLookup;
+import net.minecraft.core.HolderOwner;
 import net.minecraft.core.Registry;
 import net.minecraft.util.ExtraCodecs;
 
@@ -34,6 +_,11 @@
         this.lookupProvider = lookupProvider;
     }
 
+    protected RegistryOps(RegistryOps<T> other) {
+        super(other);
+        this.lookupProvider = other.lookupProvider;
+    }
+
     public <U> RegistryOps<U> withParent(DynamicOps<U> parent) {
         return (RegistryOps<U>)(parent == this.delegate ? this : new RegistryOps<>(parent, this.lookupProvider));
     }
@@ -71,6 +_,20 @@
             .forGetter(var0 -> null);
     }
 
+    public static <E> com.mojang.serialization.MapCodec<HolderLookup.RegistryLookup<E>> retrieveRegistryLookup(ResourceKey<? extends Registry<? extends E>> resourceKey) {
+        return ExtraCodecs.retrieveContext(ops -> {
+            if (!(ops instanceof RegistryOps<?> registryOps))
+                return DataResult.error(() -> "Not a registry ops");
+
+            return registryOps.lookupProvider.lookup(resourceKey).map(registryInfo -> {
+                if (!(registryInfo instanceof HolderLookup.RegistryLookup<E> registryLookup))
+                    return DataResult.<HolderLookup.RegistryLookup<E>>error(() -> "Found holder getter but was not a registry lookup for " + resourceKey);
+
+                return DataResult.success(registryLookup, registryLookup.registryLifecycle());
+            }).orElseGet(() -> DataResult.error(() -> "Unknown registry: " + resourceKey));
+        });
+    }
+
     public static <E, O> RecordCodecBuilder<O, Holder.Reference<E>> retrieveElement(ResourceKey<E> key) {
         ResourceKey<? extends Registry<E>> registryKey = ResourceKey.createRegistryKey(key.registry());
         return ExtraCodecs.retrieveContext(
@@ -107,6 +_,13 @@
         public int hashCode() {
             return this.lookupProvider.hashCode();
         }
+
+        // Paper start - add method to get the value for pre-filling builders in the reg mod API
+        @Override
+        public HolderLookup.Provider lookupForValueCopyViaBuilders() {
+            return this.lookupProvider;
+        }
+        // Paper end - add method to get the value for pre-filling builders in the reg mod API
     }
 
     public interface RegistryInfoLookup {
