--- a/net/minecraft/resources/ResourceLocation.java
+++ b/net/minecraft/resources/ResourceLocation.java
@@ -32,6 +_,7 @@
     public static final char NAMESPACE_SEPARATOR = ':';
     public static final String DEFAULT_NAMESPACE = "minecraft";
     public static final String REALMS_NAMESPACE = "realms";
+    public static final String PAPER_NAMESPACE = "paper"; // Paper
     private final String namespace;
     private final String path;
 
@@ -163,6 +_,12 @@
         }
 
         return i;
+    }
+
+    // Normal compare sorts by path first, this compares namespace first.
+    public int compareNamespaced(ResourceLocation o) {
+        int ret = this.namespace.compareTo(o.namespace);
+        return ret != 0 ? ret : this.path.compareTo(o.path);
     }
 
     public String toDebugFileName() {
