--- a/net/minecraft/resources/ResourceManagerRegistryLoadTask.java
+++ b/net/minecraft/resources/ResourceManagerRegistryLoadTask.java
@@ -26,21 +_,32 @@
         return new RegistrationInfo(knownPack, lifecycle);
     });
     private final ResourceManager resourceManager;
+    private final List<net.minecraft.core.Registry.PendingTags<?>> pendingTags;
 
+    /// @deprecated Neo: use [#ResourceManagerRegistryLoadTask(RegistryDataLoader.RegistryData, Lifecycle, Map, ResourceManager, List)] instead
+    @Deprecated
     public ResourceManagerRegistryLoadTask(
         RegistryDataLoader.RegistryData<T> data, Lifecycle lifecycle, Map<ResourceKey<?>, Exception> loadingErrors, ResourceManager resourceManager
     ) {
+        this(data, lifecycle, loadingErrors, resourceManager, List.of());
+    }
+
+    public ResourceManagerRegistryLoadTask(
+        RegistryDataLoader.RegistryData<T> data, Lifecycle lifecycle, Map<ResourceKey<?>, Exception> loadingErrors, ResourceManager resourceManager, List<net.minecraft.core.Registry.PendingTags<?>> pendingTags
+    ) {
         super(data, lifecycle, loadingErrors);
         this.resourceManager = resourceManager;
+        this.pendingTags = pendingTags;
     }
 
     @Override
     public CompletableFuture<?> load(RegistryOps.RegistryInfoLookup context, Executor executor) {
         FileToIdConverter lister = FileToIdConverter.registry(this.registryKey());
+        final io.papermc.paper.registry.data.util.Conversions conversions = new io.papermc.paper.registry.data.util.Conversions(context); // Paper - create conversions
         return CompletableFuture.<Map<Identifier, Resource>>supplyAsync(() -> lister.listMatchingResources(this.resourceManager), executor)
             .thenCompose(
                 registryResources -> {
-                    RegistryOps<JsonElement> ops = RegistryOps.create(JsonOps.INSTANCE, context);
+                    RegistryOps<JsonElement> ops = new net.neoforged.neoforge.common.conditions.ConditionalOps<>(RegistryOps.create(JsonOps.INSTANCE, context), net.neoforged.neoforge.common.conditions.ICondition.IContext.TAGS_INVALID);
                     return ParallelMapTransform.schedule(
                         (Map<Identifier, Resource>)registryResources,
                         (resourceId, thunk) -> {
@@ -58,11 +_,13 @@
             )
             .thenAcceptAsync(
                 loadedEntries -> {
-                    this.registerElements(loadedEntries.entrySet().stream().sorted(Entry.comparingByKey()).map(Entry::getValue));
+                    this.registerElements(loadedEntries.entrySet().stream().sorted(Entry.comparingByKey()).map(Entry::getValue), conversions);
+                    io.papermc.paper.registry.PaperRegistryAccess.instance().lockReferenceHolders(this.registryKey()); // Paper - lock reference holders
+                    io.papermc.paper.registry.PaperRegistryListenerManager.INSTANCE.runFreezeListeners(this.registryKey(), conversions); // Paper - run pre-freeze listeners
                     TagLoader.ElementLookup<Holder<T>> tagElementLookup = TagLoader.ElementLookup.fromGetters(
                         this.registryKey(), this.concurrentRegistrationGetter, this.readOnlyRegistry()
                     );
-                    Map<TagKey<T>, List<Holder<T>>> pendingTags = TagLoader.loadTagsForRegistry(this.resourceManager, this.registryKey(), tagElementLookup);
+                    Map<TagKey<T>, List<Holder<T>>> pendingTags = TagLoader.loadTagsForRegistry(this.resourceManager, this.registryKey(), tagElementLookup, this.registry, io.papermc.paper.plugin.lifecycle.event.registrar.ReloadableRegistrarEvent.Cause.INITIAL); // Paper - tag lifecycle - add registry, cause
                     this.registerTags(pendingTags);
                 },
                 executor
