--- a/net/minecraft/network/chat/ChatType.java
+++ b/net/minecraft/network/chat/ChatType.java
@@ -37,7 +_,7 @@
     public static final ResourceKey<ChatType> TEAM_MSG_COMMAND_INCOMING = create("team_msg_command_incoming");
     public static final ResourceKey<ChatType> TEAM_MSG_COMMAND_OUTGOING = create("team_msg_command_outgoing");
     public static final ResourceKey<ChatType> EMOTE_COMMAND = create("emote_command");
-
+    public static final ResourceKey<ChatType> PAPER_RAW =create("raw");
     private static ResourceKey<ChatType> create(String p_237024_) {
         return ResourceKey.create(Registries.CHAT_TYPE, ResourceLocation.withDefaultNamespace(p_237024_));
     }
@@ -64,6 +_,7 @@
             new ChatType(ChatTypeDecoration.teamMessage("chat.type.team.sent"), ChatTypeDecoration.withSender("chat.type.text.narrate"))
         );
         p_321866_.register(EMOTE_COMMAND, new ChatType(ChatTypeDecoration.withSender("chat.type.emote"), ChatTypeDecoration.withSender("chat.type.emote")));
+        p_321866_.register(PAPER_RAW, new ChatType(ChatTypeDecoration.withSender("%s"), ChatTypeDecoration.withSender("%s")));
     }
 
     public static ChatType.Bound bind(ResourceKey<ChatType> p_241279_, Entity p_241483_) {
