--- a/net/minecraft/network/chat/ComponentUtils.java
+++ b/net/minecraft/network/chat/ComponentUtils.java
@@ -43,6 +_,11 @@
         if (p_130735_ > 100) {
             return p_130733_.copy();
         } else {
+            // Paper start - adventure; pass actual vanilla component
+            if (p_130733_ instanceof io.papermc.paper.adventure.AdventureComponent adventureComponent) {
+                p_130733_ = adventureComponent.deepConverted();
+            }
+            // Paper end - adventure; pass actual vanilla component
             MutableComponent mutablecomponent = p_130733_.getContents().resolve(p_130732_, p_130734_, p_130735_ + 1);
 
             for (Component component : p_130733_.getSiblings()) {
