--- a/net/minecraft/network/chat/FilterMask.java
+++ b/net/minecraft/network/chat/FilterMask.java
@@ -1,16 +_,11 @@
 package net.minecraft.network.chat;
 
-import com.google.common.base.Suppliers;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
-import io.netty.buffer.ByteBuf;
 import java.util.BitSet;
-import java.util.function.IntFunction;
 import java.util.function.Supplier;
 import net.minecraft.ChatFormatting;
-import net.minecraft.network.codec.ByteBufCodecs;
-import net.minecraft.network.codec.StreamCodec;
-import net.minecraft.util.ByIdMap;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.StringRepresentable;
 import org.apache.commons.lang3.StringUtils;
@@ -18,12 +_,14 @@
 
 public class FilterMask {
     public static final Codec<FilterMask> CODEC = StringRepresentable.fromEnum(FilterMask.Type::values).dispatch(FilterMask::type, FilterMask.Type::codec);
-    public static final StreamCodec<ByteBuf, FilterMask> STREAM_CODEC = FilterMask.Type.STREAM_CODEC.dispatch(FilterMask::type, FilterMask.Type::streamCodec);
     public static final FilterMask FULLY_FILTERED = new FilterMask(new BitSet(0), FilterMask.Type.FULLY_FILTERED);
     public static final FilterMask PASS_THROUGH = new FilterMask(new BitSet(0), FilterMask.Type.PASS_THROUGH);
     public static final Style FILTERED_STYLE = Style.EMPTY
         .withColor(ChatFormatting.DARK_GRAY)
         .withHoverEvent(new HoverEvent.ShowText(Component.translatable("chat.filtered")));
+    private static final MapCodec<FilterMask> PASS_THROUGH_CODEC = MapCodec.unit(PASS_THROUGH);
+    private static final MapCodec<FilterMask> FULLY_FILTERED_CODEC = MapCodec.unit(FULLY_FILTERED);
+    private static final MapCodec<FilterMask> PARTIALLY_FILTERED_CODEC = ExtraCodecs.BIT_SET.xmap(FilterMask::new, FilterMask::mask).fieldOf("value");
     private static final char HASH = '#';
     private final BitSet mask;
     private final FilterMask.Type type;
@@ -50,6 +_,23 @@
         return this.mask;
     }
 
+    public static FilterMask read(FriendlyByteBuf input) {
+        FilterMask.Type type = input.readEnum(FilterMask.Type.class);
+
+        return switch (type) {
+            case PASS_THROUGH -> PASS_THROUGH;
+            case FULLY_FILTERED -> FULLY_FILTERED;
+            case PARTIALLY_FILTERED -> new FilterMask(input.readBitSet(), FilterMask.Type.PARTIALLY_FILTERED);
+        };
+    }
+
+    public static void write(FriendlyByteBuf output, FilterMask mask) {
+        output.writeEnum(mask.type);
+        if (mask.type == FilterMask.Type.PARTIALLY_FILTERED) {
+            output.writeBitSet(mask.mask);
+        }
+    }
+
     public void setFiltered(int index) {
         this.mask.set(index);
     }
@@ -128,27 +_,16 @@
     }
 
     private enum Type implements StringRepresentable {
-        PASS_THROUGH(0, "pass_through", () -> MapCodec.unit(FilterMask.PASS_THROUGH), () -> StreamCodec.unit(FilterMask.PASS_THROUGH)),
-        FULLY_FILTERED(1, "fully_filtered", () -> MapCodec.unit(FilterMask.FULLY_FILTERED), () -> StreamCodec.unit(FilterMask.FULLY_FILTERED)),
-        PARTIALLY_FILTERED(
-            2,
-            "partially_filtered",
-            () -> ExtraCodecs.BIT_SET.xmap(FilterMask::new, FilterMask::mask).fieldOf("value"),
-            () -> ByteBufCodecs.BIT_SET.map(FilterMask::new, FilterMask::mask)
-        );
+        PASS_THROUGH("pass_through", () -> FilterMask.PASS_THROUGH_CODEC),
+        FULLY_FILTERED("fully_filtered", () -> FilterMask.FULLY_FILTERED_CODEC),
+        PARTIALLY_FILTERED("partially_filtered", () -> FilterMask.PARTIALLY_FILTERED_CODEC);
 
-        private static final IntFunction<FilterMask.Type> ID_MAP = ByIdMap.continuous(t -> t.id, values(), ByIdMap.OutOfBoundsStrategy.ZERO);
-        public static final StreamCodec<ByteBuf, FilterMask.Type> STREAM_CODEC = ByteBufCodecs.idMapper(ID_MAP, t -> t.id);
-        private final int id;
         private final String serializedName;
         private final Supplier<MapCodec<FilterMask>> codec;
-        private final Supplier<StreamCodec<ByteBuf, FilterMask>> streamCodec;
 
-        Type(int id, String serializedName, Supplier<MapCodec<FilterMask>> codec, Supplier<StreamCodec<ByteBuf, FilterMask>> streamCodec) {
-            this.id = id;
+        Type(String serializedName, Supplier<MapCodec<FilterMask>> codec) {
             this.serializedName = serializedName;
-            this.codec = Suppliers.memoize(codec::get);
-            this.streamCodec = Suppliers.memoize(streamCodec::get);
+            this.codec = codec;
         }
 
         @Override
@@ -158,10 +_,6 @@
 
         private MapCodec<FilterMask> codec() {
             return this.codec.get();
-        }
-
-        private StreamCodec<ByteBuf, FilterMask> streamCodec() {
-            return this.streamCodec.get();
         }
     }
 }
