--- a/net/minecraft/network/chat/HoverEvent.java
+++ b/net/minecraft/network/chat/HoverEvent.java
@@ -225,8 +_,8 @@
         private static final Codec<HoverEvent.ItemStackInfo> SIMPLE_CODEC = ItemStack.SIMPLE_ITEM_CODEC
             .xmap(HoverEvent.ItemStackInfo::new, HoverEvent.ItemStackInfo::getItemStack);
         public static final Codec<HoverEvent.ItemStackInfo> CODEC = Codec.withAlternative(FULL_CODEC, SIMPLE_CODEC);
-        private final Holder<Item> item;
-        private final int count;
+        public final Holder<Item> item;
+        public final int count;
         private final DataComponentPatch components;
         @Nullable
         private ItemStack itemStack;
