--- a/net/minecraft/network/chat/LastSeenMessagesValidator.java
+++ b/net/minecraft/network/chat/LastSeenMessagesValidator.java
@@ -6,7 +_,7 @@
 
 public class LastSeenMessagesValidator {
     private final int lastSeenCount;
-    private final ObjectList<@Nullable LastSeenTrackedEntry> trackedMessages = new ObjectArrayList<>();
+    private final ObjectList<LastSeenTrackedEntry> trackedMessages = new ObjectArrayList<>();
     private @Nullable MessageSignature lastPendingMessage;
 
     public LastSeenMessagesValidator(int lastSeenCount) {
