--- a/net/minecraft/network/chat/LastSeenMessages.java
+++ b/net/minecraft/network/chat/LastSeenMessages.java
@@ -2,14 +_,12 @@
 
 import com.google.common.primitives.Ints;
 import com.mojang.serialization.Codec;
-import io.netty.buffer.ByteBuf;
 import java.security.SignatureException;
 import java.util.ArrayList;
 import java.util.BitSet;
 import java.util.List;
 import java.util.Optional;
-import net.minecraft.network.codec.ByteBufCodecs;
-import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.util.SignatureUpdater;
 
 public record LastSeenMessages(List<MessageSignature> entries) {
@@ -42,9 +_,14 @@
 
     public record Packed(List<MessageSignature.Packed> entries) {
         public static final LastSeenMessages.Packed EMPTY = new LastSeenMessages.Packed(List.of());
-        public static final StreamCodec<ByteBuf, LastSeenMessages.Packed> STREAM_CODEC = StreamCodec.composite(
-            MessageSignature.Packed.STREAM_CODEC.apply(ByteBufCodecs.list(20)), LastSeenMessages.Packed::entries, LastSeenMessages.Packed::new
-        );
+
+        public Packed(FriendlyByteBuf input) {
+            this(input.readCollection(FriendlyByteBuf.<List<MessageSignature.Packed>>limitValue(ArrayList::new, 20), MessageSignature.Packed::read));
+        }
+
+        public void write(FriendlyByteBuf output) {
+            output.writeCollection(this.entries, MessageSignature.Packed::write);
+        }
 
         public Optional<LastSeenMessages> unpack(MessageSignatureCache cache) {
             List<MessageSignature> unpacked = new ArrayList<>(this.entries.size());
@@ -64,15 +_,16 @@
 
     public record Update(int offset, BitSet acknowledged, byte checksum) {
         public static final byte IGNORE_CHECKSUM = 0;
-        public static final StreamCodec<ByteBuf, LastSeenMessages.Update> STREAM_CODEC = StreamCodec.composite(
-            ByteBufCodecs.VAR_INT,
-            LastSeenMessages.Update::offset,
-            ByteBufCodecs.fixedBitSet(20),
-            LastSeenMessages.Update::acknowledged,
-            ByteBufCodecs.BYTE,
-            LastSeenMessages.Update::checksum,
-            LastSeenMessages.Update::new
-        );
+
+        public Update(FriendlyByteBuf input) {
+            this(input.readVarInt(), input.readFixedBitSet(20), input.readByte());
+        }
+
+        public void write(FriendlyByteBuf output) {
+            output.writeVarInt(this.offset);
+            output.writeFixedBitSet(this.acknowledged, 20);
+            output.writeByte(this.checksum);
+        }
 
         public boolean verifyChecksum(LastSeenMessages lastSeen) {
             return this.checksum == 0 || this.checksum == lastSeen.computeChecksum();
