--- a/net/minecraft/network/chat/RemoteChatSession.java
+++ b/net/minecraft/network/chat/RemoteChatSession.java
@@ -1,11 +_,9 @@
 package net.minecraft.network.chat;
 
 import com.mojang.authlib.GameProfile;
-import io.netty.buffer.ByteBuf;
 import java.time.Duration;
 import java.util.UUID;
-import net.minecraft.core.UUIDUtil;
-import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.util.SignatureValidator;
 import net.minecraft.world.entity.player.ProfilePublicKey;
 
@@ -27,13 +_,14 @@
     }
 
     public record Data(UUID sessionId, ProfilePublicKey.Data profilePublicKey) {
-        public static final StreamCodec<ByteBuf, RemoteChatSession.Data> STREAM_CODEC = StreamCodec.composite(
-            UUIDUtil.STREAM_CODEC,
-            RemoteChatSession.Data::sessionId,
-            ProfilePublicKey.Data.STREAM_CODEC,
-            RemoteChatSession.Data::profilePublicKey,
-            RemoteChatSession.Data::new
-        );
+        public static RemoteChatSession.Data read(FriendlyByteBuf input) {
+            return new RemoteChatSession.Data(input.readUUID(), new ProfilePublicKey.Data(input));
+        }
+
+        public static void write(FriendlyByteBuf output, RemoteChatSession.Data data) {
+            output.writeUUID(data.sessionId);
+            data.profilePublicKey.write(output);
+        }
 
         public RemoteChatSession validate(GameProfile profile, SignatureValidator serviceSignatureValidator) throws ProfilePublicKey.ValidationException {
             return new RemoteChatSession(this.sessionId, ProfilePublicKey.createValidated(serviceSignatureValidator, profile.id(), this.profilePublicKey));
