--- a/net/minecraft/network/chat/TextColor.java
+++ b/net/minecraft/network/chat/TextColor.java
@@ -23,7 +_,7 @@
         .collect(ImmutableMap.toImmutableMap(p_237297_ -> p_237297_.name, Function.identity()));
     private final int value;
     @Nullable
-    public final String name;
+    private final String name;
 
     private TextColor(int p_131263_, String p_131264_) {
         this.value = p_131263_ & 16777215;
