--- a/net/minecraft/world/level/block/entity/vault/VaultServerData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
@@ -74,11 +_,11 @@
         this.markChanged();
     }
 
-    public long stateUpdatingResumesAt() {
+    long stateUpdatingResumesAt() {
         return this.stateUpdatingResumesAt;
     }
 
-    public void pauseStateUpdatingUntil(long stateUpdatingResumesAt) {
+    void pauseStateUpdatingUntil(long stateUpdatingResumesAt) {
         this.stateUpdatingResumesAt = stateUpdatingResumesAt;
         this.markChanged();
     }
