--- a/net/minecraft/world/level/levelgen/blockpredicates/BlockPredicate.java
+++ b/net/minecraft/world/level/levelgen/blockpredicates/BlockPredicate.java
@@ -101,10 +_,6 @@
         return new HasSturdyFacePredicate(Vec3i.ZERO, direction);
     }
 
-    static BlockPredicate hasSturdyFace(Directional offset, Direction direction) {
-        return new HasSturdyFacePredicate(offset.getStep(), direction);
-    }
-
     static BlockPredicate solid(Directional directional) {
         return new SolidPredicate(directional.getStep());
     }
@@ -131,9 +_,5 @@
 
     static BlockPredicate heightRange(VerticalAnchor minInclusive, VerticalAnchor maxInclusive) {
         return new HeightRangePredicate(minInclusive, maxInclusive);
-    }
-
-    static BlockPredicate volumeMatch(Vec3i min, Vec3i max, BlockPredicate match) {
-        return new VolumeMatchPredicate(min, max, match);
     }
 }
