--- a/net/minecraft/world/level/levelgen/blockpredicates/BlockPredicateType.java
+++ b/net/minecraft/world/level/levelgen/blockpredicates/BlockPredicateType.java
@@ -20,7 +_,6 @@
     BlockPredicateType<TrueBlockPredicate> TRUE = register("true", TrueBlockPredicate.CODEC);
     BlockPredicateType<UnobstructedPredicate> UNOBSTRUCTED = register("unobstructed", UnobstructedPredicate.CODEC);
     BlockPredicateType<HeightRangePredicate> HEIGHT_RANGE = register("height_range", HeightRangePredicate.CODEC);
-    BlockPredicateType<VolumeMatchPredicate> VOLUME_MATCH = register("volume_match", VolumeMatchPredicate.CODEC);
 
     MapCodec<P> codec();
 
