--- a/net/minecraft/world/level/levelgen/blockpredicates/MatchingBiomesPredicate.java
+++ b/net/minecraft/world/level/levelgen/blockpredicates/MatchingBiomesPredicate.java
@@ -4,14 +_,14 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.biome.Biome;
 
 public class MatchingBiomesPredicate implements BlockPredicate {
     public static final MapCodec<MatchingBiomesPredicate> CODEC = RecordCodecBuilder.mapCodec(
-        i -> i.group(RegistryCodecs.holderSet(Registries.BIOME).fieldOf("biomes").forGetter(c -> c.biomes)).apply(i, MatchingBiomesPredicate::new)
+        i -> i.group(RegistryCodecs.homogeneousList(Registries.BIOME).fieldOf("biomes").forGetter(c -> c.biomes)).apply(i, MatchingBiomesPredicate::new)
     );
     private final HolderSet<Biome> biomes;
 
