--- a/net/minecraft/world/level/levelgen/blockpredicates/MatchingFluidsPredicate.java
+++ b/net/minecraft/world/level/levelgen/blockpredicates/MatchingFluidsPredicate.java
@@ -3,16 +_,16 @@
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.Vec3i;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluid;
 
 public class MatchingFluidsPredicate extends StateTestingPredicate {
     public static final MapCodec<MatchingFluidsPredicate> CODEC = RecordCodecBuilder.mapCodec(
         i -> stateTestingCodec(i)
-            .and(RegistryCodecs.holderSet(Registries.FLUID).fieldOf("fluids").forGetter(c -> c.fluids))
+            .and(RegistryCodecs.homogeneousList(Registries.FLUID).fieldOf("fluids").forGetter(c -> c.fluids))
             .apply(i, MatchingFluidsPredicate::new)
     );
     private final HolderSet<Fluid> fluids;
