--- a/net/minecraft/client/renderer/BindGroupLayouts.java
+++ b/net/minecraft/client/renderer/BindGroupLayouts.java
@@ -11,47 +_,33 @@
     public static final BindGroupLayout DYNAMIC_TRANSFORMS = BindGroupLayout.builder().withUniform("DynamicTransforms", UniformType.UNIFORM_BUFFER).build();
     public static final BindGroupLayout PROJECTION = BindGroupLayout.builder().withUniform("Projection", UniformType.UNIFORM_BUFFER).build();
     public static final BindGroupLayout CHUNK_SECTION = BindGroupLayout.builder().withUniform("ChunkSection", UniformType.UNIFORM_BUFFER).build();
-    public static final BindGroupLayout TERRAIN_INFO = BindGroupLayout.builder().withUniform("TerrainUniform", UniformType.UNIFORM_BUFFER).build();
     public static final BindGroupLayout FOG = BindGroupLayout.builder().withUniform("Fog", UniformType.UNIFORM_BUFFER).build();
     public static final BindGroupLayout GLOBALS = BindGroupLayout.builder().withUniform("Globals", UniformType.UNIFORM_BUFFER).build();
     public static final BindGroupLayout LIGHTING = BindGroupLayout.builder().withUniform("Lighting", UniformType.UNIFORM_BUFFER).build();
-    public static final BindGroupLayout SAMPLER0 = BindGroupLayout.builder().withUniform("Sampler0", UniformType.COMBINED_IMAGE_SAMPLER).build();
-    public static final BindGroupLayout SAMPLER1 = BindGroupLayout.builder().withUniform("Sampler1", UniformType.COMBINED_IMAGE_SAMPLER).build();
-    public static final BindGroupLayout SAMPLER2 = BindGroupLayout.builder().withUniform("Sampler2", UniformType.COMBINED_IMAGE_SAMPLER).build();
-    public static final BindGroupLayout SAMPLER0_SAMPLER2 = BindGroupLayout.builder()
-        .withUniform("Sampler0", UniformType.COMBINED_IMAGE_SAMPLER)
-        .withUniform("Sampler2", UniformType.COMBINED_IMAGE_SAMPLER)
-        .build();
-    public static final BindGroupLayout SAMPLER0_SAMPLER1 = BindGroupLayout.builder()
-        .withUniform("Sampler0", UniformType.COMBINED_IMAGE_SAMPLER)
-        .withUniform("Sampler1", UniformType.COMBINED_IMAGE_SAMPLER)
-        .build();
+    public static final BindGroupLayout SAMPLER0 = BindGroupLayout.builder().withSampler("Sampler0").build();
+    public static final BindGroupLayout SAMPLER1 = BindGroupLayout.builder().withSampler("Sampler1").build();
+    public static final BindGroupLayout SAMPLER2 = BindGroupLayout.builder().withSampler("Sampler2").build();
+    public static final BindGroupLayout SAMPLER0_SAMPLER2 = BindGroupLayout.builder().withSampler("Sampler0").withSampler("Sampler2").build();
+    public static final BindGroupLayout SAMPLER0_SAMPLER1 = BindGroupLayout.builder().withSampler("Sampler0").withSampler("Sampler1").build();
     public static final BindGroupLayout SAMPLER0_SAMPLER1_SAMPLER2 = BindGroupLayout.builder()
-        .withUniform("Sampler0", UniformType.COMBINED_IMAGE_SAMPLER)
-        .withUniform("Sampler1", UniformType.COMBINED_IMAGE_SAMPLER)
-        .withUniform("Sampler2", UniformType.COMBINED_IMAGE_SAMPLER)
+        .withSampler("Sampler0")
+        .withSampler("Sampler1")
+        .withSampler("Sampler2")
         .build();
     public static final BindGroupLayout CLOUD_INFO = BindGroupLayout.builder()
         .withUniform("CloudInfo", UniformType.UNIFORM_BUFFER)
         .withUniform("CloudFaces", UniformType.TEXEL_BUFFER, GpuFormat.R8_SINT)
         .build();
-    public static final BindGroupLayout DISSOLVE_MASK_SAMPLER = BindGroupLayout.builder()
-        .withUniform("DissolveMaskSampler", UniformType.COMBINED_IMAGE_SAMPLER)
-        .build();
-    public static final BindGroupLayout IN_SAMPLER = BindGroupLayout.builder().withUniform("InSampler", UniformType.COMBINED_IMAGE_SAMPLER).build();
+    public static final BindGroupLayout DISSOLVE_MASK_SAMPLER = BindGroupLayout.builder().withSampler("DissolveMaskSampler").build();
+    public static final BindGroupLayout IN_SAMPLER = BindGroupLayout.builder().withSampler("InSampler").build();
     public static final BindGroupLayout LIGHTMAP_INFO = BindGroupLayout.builder().withUniform("LightmapInfo", UniformType.UNIFORM_BUFFER).build();
     public static final BindGroupLayout SPRITE_ANIMATION_INFO = BindGroupLayout.builder()
         .withUniform("SpriteAnimationInfo", UniformType.UNIFORM_BUFFER)
         .build();
-    public static final BindGroupLayout SPRITE = BindGroupLayout.builder().withUniform("Sprite", UniformType.COMBINED_IMAGE_SAMPLER).build();
-    public static final BindGroupLayout CURRENT_SPRITE_NEXT_SPRITE = BindGroupLayout.builder()
-        .withUniform("CurrentSprite", UniformType.COMBINED_IMAGE_SAMPLER)
-        .withUniform("NextSprite", UniformType.COMBINED_IMAGE_SAMPLER)
-        .build();
-    public static final BindGroupLayout GLINT_SAMPLER = BindGroupLayout.builder().withUniform("GlintSampler", UniformType.COMBINED_IMAGE_SAMPLER).build();
-    public static final BindGroupLayout DEPTH_BOUNDS_SAMPLER = BindGroupLayout.builder()
-        .withUniform("DepthBoundsSampler", UniformType.COMBINED_IMAGE_SAMPLER)
-        .build();
+    public static final BindGroupLayout SPRITE = BindGroupLayout.builder().withSampler("Sprite").build();
+    public static final BindGroupLayout CURRENT_SPRITE_NEXT_SPRITE = BindGroupLayout.builder().withSampler("CurrentSprite").withSampler("NextSprite").build();
+    public static final BindGroupLayout GLINT_SAMPLER = BindGroupLayout.builder().withSampler("GlintSampler").build();
+    public static final BindGroupLayout DEPTH_BOUNDS_SAMPLER = BindGroupLayout.builder().withSampler("DepthBoundsSampler").build();
     public static final BindGroupLayout OIT_COEFFS_DEPTH_BOUNDS_SAMPLER;
     public static final BindGroupLayout SAMPLER0_OIT_COEFFS_DEPTH_BOUNDS_SAMPLER;
     public static final BindGroupLayout SAMPLER0_SAMPLER2_OIT_COEFFS_DEPTH_BOUNDS_SAMPLER;
@@ -61,39 +_,34 @@
     }
 
     static {
-        BindGroupLayout.Builder builder = BindGroupLayout.builder().withUniform("DepthBoundsSampler", UniformType.COMBINED_IMAGE_SAMPLER);
+        BindGroupLayout.Builder builder = BindGroupLayout.builder().withSampler("DepthBoundsSampler");
 
         for (int i = 0; i < LevelRenderer.OIT_TRANSMITTANCE_TARGET_COUNT; i++) {
-            builder.withUniform("Coeff" + i, UniformType.COMBINED_IMAGE_SAMPLER);
+            builder.withSampler("Coeff" + i);
         }
 
         OIT_COEFFS_DEPTH_BOUNDS_SAMPLER = builder.build();
-        builder = BindGroupLayout.builder()
-            .withUniform("Sampler0", UniformType.COMBINED_IMAGE_SAMPLER)
-            .withUniform("DepthBoundsSampler", UniformType.COMBINED_IMAGE_SAMPLER);
+        builder = BindGroupLayout.builder().withSampler("DepthBoundsSampler").withSampler("Sampler0");
 
         for (int i = 0; i < LevelRenderer.OIT_TRANSMITTANCE_TARGET_COUNT; i++) {
-            builder.withUniform("Coeff" + i, UniformType.COMBINED_IMAGE_SAMPLER);
+            builder.withSampler("Coeff" + i);
         }
 
         SAMPLER0_OIT_COEFFS_DEPTH_BOUNDS_SAMPLER = builder.build();
-        builder = BindGroupLayout.builder()
-            .withUniform("Sampler0", UniformType.COMBINED_IMAGE_SAMPLER)
-            .withUniform("Sampler2", UniformType.COMBINED_IMAGE_SAMPLER)
-            .withUniform("DepthBoundsSampler", UniformType.COMBINED_IMAGE_SAMPLER);
+        builder = BindGroupLayout.builder().withSampler("Sampler0").withSampler("Sampler2").withSampler("DepthBoundsSampler");
 
         for (int i = 0; i < LevelRenderer.OIT_TRANSMITTANCE_TARGET_COUNT; i++) {
-            builder.withUniform("Coeff" + i, UniformType.COMBINED_IMAGE_SAMPLER);
+            builder.withSampler("Coeff" + i);
         }
 
         SAMPLER0_SAMPLER2_OIT_COEFFS_DEPTH_BOUNDS_SAMPLER = builder.build();
         builder = BindGroupLayout.builder()
             .withUniform("CloudInfo", UniformType.UNIFORM_BUFFER)
             .withUniform("CloudFaces", UniformType.TEXEL_BUFFER, GpuFormat.R8_SINT)
-            .withUniform("DepthBoundsSampler", UniformType.COMBINED_IMAGE_SAMPLER);
+            .withSampler("DepthBoundsSampler");
 
         for (int i = 0; i < LevelRenderer.OIT_TRANSMITTANCE_TARGET_COUNT; i++) {
-            builder.withUniform("Coeff" + i, UniformType.COMBINED_IMAGE_SAMPLER);
+            builder.withSampler("Coeff" + i);
         }
 
         CLOUD_INFO_OIT_COEFFS_DEPTH_BOUNDS_SAMPLER = builder.build();
