--- a/net/minecraft/client/renderer/GlobalSettingsUniform.java
+++ b/net/minecraft/client/renderer/GlobalSettingsUniform.java
@@ -13,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class GlobalSettingsUniform implements AutoCloseable {
-    public static final int UBO_SIZE = new Std140SizeCalculator().putIVec3().putFloat().putVec3().putFloat().putVec2().putInt().putInt().get();
+    public static final int UBO_SIZE = new Std140SizeCalculator().putIVec3().putVec3().putVec2().putFloat().putFloat().putInt().putInt().get();
     private final GpuBuffer buffer = RenderSystem.getDevice().createBuffer(() -> "Global Settings UBO", 136, UBO_SIZE);
 
     public void update(int width, int height, double glintAlpha, long gameTime, float worldPartialTicks, int menuBlurRadius, Vec3 cameraPos, boolean useRgss) {
@@ -23,10 +_,10 @@
             int cameraZ = Mth.floor(cameraPos.z);
             ByteBuffer data = Std140Builder.onStack(stack, UBO_SIZE)
                 .putIVec3(cameraX, cameraY, cameraZ)
-                .putFloat((float)glintAlpha)
                 .putVec3((float)(cameraX - cameraPos.x), (float)(cameraY - cameraPos.y), (float)(cameraZ - cameraPos.z))
-                .putFloat(((float)(gameTime % 24000L) + worldPartialTicks) / 24000.0F)
                 .putVec2(width, height)
+                .putFloat((float)glintAlpha)
+                .putFloat(((float)(gameTime % 24000L) + worldPartialTicks) / 24000.0F)
                 .putInt(menuBlurRadius)
                 .putInt(useRgss ? 1 : 0)
                 .get();
