--- a/net/minecraft/client/renderer/LevelTargetBundle.java
+++ b/net/minecraft/client/renderer/LevelTargetBundle.java
@@ -19,7 +_,7 @@
     public ResourceHandle<RenderTarget> main = ResourceHandle.invalid();
     public ResourceHandle<RenderTarget> alwaysOnTopDepth = ResourceHandle.invalid();
     public ResourceHandle<RenderTarget> depthBounds = ResourceHandle.invalid();
-    public ResourceHandle<RenderTarget> depthBoundsCulled = ResourceHandle.invalid();
+    public ResourceHandle<RenderTarget> depthBoundsCopy = ResourceHandle.invalid();
     public final List<ResourceHandle<RenderTarget>> transmittance = new ArrayList<>();
     public ResourceHandle<RenderTarget> accumulate = ResourceHandle.invalid();
     public ResourceHandle<RenderTarget> oitCloudDepth = ResourceHandle.invalid();
