--- a/net/minecraft/client/renderer/LevelRenderer.java
+++ b/net/minecraft/client/renderer/LevelRenderer.java
@@ -116,7 +_,7 @@
     private final TextureManager textureManager;
     private final AtlasManager atlasManager;
     private final ShaderManager shaderManager;
-    final LevelRenderState levelRenderState;
+    private final LevelRenderState levelRenderState;
     private final OptionsRenderState optionsRenderState;
     private @Nullable SkyRenderer skyRenderer;
     private final CloudRenderer cloudRenderer = new CloudRenderer();
