--- a/net/minecraft/client/renderer/LightmapRenderStateExtractor.java
+++ b/net/minecraft/client/renderer/LightmapRenderStateExtractor.java
@@ -5,6 +_,7 @@
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.client.player.LocalPlayer;
 import net.minecraft.client.renderer.state.LightmapRenderState;
+import net.minecraft.util.ARGB;
 import net.minecraft.util.Mth;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.profiling.Profiler;
@@ -53,9 +_,9 @@
                 profiler.push("lightmap");
                 Camera camera = this.renderer.mainCamera();
                 renderState.blockFactor = this.blockLightFlicker + 1.4F;
-                renderState.blockLightTint = camera.attributeProbe().getValue(EnvironmentAttributes.BLOCK_LIGHT_TINT, partialTicks);
+                renderState.blockLightTint = ARGB.vector3fFromRGB24(camera.attributeProbe().getValue(EnvironmentAttributes.BLOCK_LIGHT_TINT, partialTicks));
                 renderState.skyFactor = camera.attributeProbe().getValue(EnvironmentAttributes.SKY_LIGHT_FACTOR, partialTicks);
-                renderState.skyLightColor = camera.attributeProbe().getValue(EnvironmentAttributes.SKY_LIGHT_COLOR, partialTicks);
+                renderState.skyLightColor = ARGB.vector3fFromRGB24(camera.attributeProbe().getValue(EnvironmentAttributes.SKY_LIGHT_COLOR, partialTicks));
                 EndFlashState endFlashState = level.endFlashState();
                 if (endFlashState != null && !this.minecraft.options.hideLightningFlash().get()) {
                     float intensity = endFlashState.getIntensity(partialTicks);
@@ -66,7 +_,7 @@
                     }
                 }
 
-                renderState.ambientColor = camera.attributeProbe().getValue(EnvironmentAttributes.AMBIENT_LIGHT_COLOR, partialTicks);
+                renderState.ambientColor = ARGB.vector3fFromRGB24(camera.attributeProbe().getValue(EnvironmentAttributes.AMBIENT_LIGHT_COLOR, partialTicks));
                 float brightnessOption = this.minecraft.options.gamma().get().floatValue();
                 float darknessEffectScaleOption = this.minecraft.options.darknessEffectScale().get().floatValue();
                 float darknessEffectBrightnessModifier = player.getEffectBlendFactor(MobEffects.DARKNESS, partialTicks) * darknessEffectScaleOption;
@@ -82,7 +_,7 @@
                     renderState.nightVisionEffectIntensity = 0.0F;
                 }
 
-                renderState.nightVisionColor = camera.attributeProbe().getValue(EnvironmentAttributes.NIGHT_VISION_COLOR, partialTicks);
+                renderState.nightVisionColor = ARGB.vector3fFromRGB24(camera.attributeProbe().getValue(EnvironmentAttributes.NIGHT_VISION_COLOR, partialTicks));
                 renderState.bossOverlayWorldDarkening = this.renderer.bossOverlayWorldDarkening(partialTicks);
                 profiler.pop();
                 this.needsUpdate = false;
