--- a/net/minecraft/client/renderer/MappableRingBuffer.java
+++ b/net/minecraft/client/renderer/MappableRingBuffer.java
@@ -39,7 +_,7 @@
     public GpuBuffer currentBuffer() {
         GpuFence fence = this.fences[this.current];
         if (fence != null) {
-            fence.awaitCompletion(-1L);
+            fence.awaitCompletion(Long.MAX_VALUE);
             fence.close();
             this.fences[this.current] = null;
         }
