--- a/net/minecraft/client/renderer/MapRenderer.java
+++ b/net/minecraft/client/renderer/MapRenderer.java
@@ -45,7 +_,7 @@
             if (!showOnlyFrame || decoration.renderOnFrame) {
                 poseStack.pushPose();
                 poseStack.translate(decoration.x / 2.0F + 64.0F, decoration.y / 2.0F + 64.0F, -0.02F);
-                poseStack.rotateDegrees(Axis.ZP, decoration.rot * 360 / 16.0F);
+                poseStack.mulPose(Axis.ZP.rotationDegrees(decoration.rot * 360 / 16.0F));
                 poseStack.scale(4.0F, 4.0F, 3.0F);
                 poseStack.translate(-0.125F, 0.125F, 0.0F);
                 TextureAtlasSprite atlasSprite = decoration.atlasSprite;
