--- a/net/minecraft/client/renderer/OrderedSubmitNodeCollector.java
+++ b/net/minecraft/client/renderer/OrderedSubmitNodeCollector.java
@@ -57,8 +_,6 @@
         int outlineColor
     );
 
-    void submitTextBackground(PoseStack poseStack, float x0, float y0, float x1, float y1, int color, Font.DisplayMode displayMode, int lightCoords);
-
     void submitFlame(PoseStack poseStack, EntityRenderState renderState, Quaternionf rotation);
 
     void submitLeash(PoseStack poseStack, EntityRenderState.LeashState leashState);
