--- a/net/minecraft/client/renderer/OrderedSubmitNodeCollector.java
+++ b/net/minecraft/client/renderer/OrderedSubmitNodeCollector.java
@@ -30,10 +_,8 @@
 import org.joml.Quaternionf;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.extensions.OrderedSubmitNodeCollectorExtension;
-
 @OnlyIn(Dist.CLIENT)
-public interface OrderedSubmitNodeCollector extends OrderedSubmitNodeCollectorExtension {
+public interface OrderedSubmitNodeCollector {
     void submitShadow(PoseStack poseStack, float radius, List<EntityRenderState.ShadowPiece> pieces);
 
     void submitNameTag(
