--- a/net/minecraft/client/renderer/PostChain.java
+++ b/net/minecraft/client/renderer/PostChain.java
@@ -70,7 +_,9 @@
         Projection projection,
         ProjectionMatrixBuffer projectionMatrixBuffer
     ) throws ShaderManager.CompilationException {
-        Set<Identifier> referencedExternalTargets = getReferencedExternalTargets(config);
+        Stream<Identifier> referencedTargets = config.passes().stream().flatMap(PostChainConfig.Pass::referencedTargets);
+        Set<Identifier> referencedExternalTargets = referencedTargets.filter(targetId -> !config.internalTargets().containsKey(targetId))
+            .collect(Collectors.toSet());
         Set<Identifier> invalidExternalTargets = Sets.difference(referencedExternalTargets, allowedExternalTargets);
         if (!invalidExternalTargets.isEmpty()) {
             throw new ShaderManager.CompilationException("Referenced external targets are not available in this context: " + invalidExternalTargets);
@@ -86,11 +_,6 @@
         return new PostChain(id, passes.build(), config.internalTargets(), referencedExternalTargets, projection, projectionMatrixBuffer);
     }
 
-    public static Set<Identifier> getReferencedExternalTargets(PostChainConfig config) {
-        Stream<Identifier> referencedTargets = config.passes().stream().flatMap(PostChainConfig.Pass::referencedTargets);
-        return referencedTargets.filter(targetId -> !config.internalTargets().containsKey(targetId)).collect(Collectors.toSet());
-    }
-
     private static PostPass createPass(TextureManager textureManager, PostChainConfig.Pass config, Identifier id) throws ShaderManager.CompilationException {
         RenderPipeline.Builder pipelineBuilder = RenderPipeline.builder(RenderPipelines.POST_PROCESSING_SNIPPET)
             .withFragmentShader(config.fragmentShaderId())
@@ -99,7 +_,7 @@
         BindGroupLayout.Builder bindGroupLayoutBuilder = BindGroupLayout.builder();
 
         for (PostChainConfig.Input input : config.inputs()) {
-            bindGroupLayoutBuilder.withUniform(input.samplerName() + "Sampler", UniformType.COMBINED_IMAGE_SAMPLER);
+            bindGroupLayoutBuilder.withSampler(input.samplerName() + "Sampler");
         }
 
         bindGroupLayoutBuilder.withUniform("SamplerInfo", UniformType.UNIFORM_BUFFER);
@@ -192,14 +_,10 @@
         return target;
     }
 
-    public void closePersistentTargets() {
-        this.persistentTargets.values().forEach(RenderTarget::destroyBuffers);
-        this.persistentTargets.clear();
-    }
-
     @Override
     public void close() {
-        this.closePersistentTargets();
+        this.persistentTargets.values().forEach(RenderTarget::destroyBuffers);
+        this.persistentTargets.clear();
 
         for (PostPass pass : this.passes) {
             pass.close();
