--- a/net/minecraft/client/renderer/PostPass.java
+++ b/net/minecraft/client/renderer/PostPass.java
@@ -127,7 +_,7 @@
                     }
 
                     for (PostPass.InputTexture inputx : inputTextures) {
-                        renderPass.setUniform(inputx.samplerName() + "Sampler", inputx.view(), inputx.sampler());
+                        renderPass.bindTexture(inputx.samplerName() + "Sampler", inputx.view(), inputx.sampler());
                     }
 
                     renderPass.draw(3, 1, 0, 0);
