--- a/net/minecraft/client/renderer/RenderPipelines.java
+++ b/net/minecraft/client/renderer/RenderPipelines.java
@@ -24,13 +_,13 @@
     private static final Map<Identifier, RenderPipeline> PIPELINES_BY_LOCATION = new HashMap<>();
     private static final float ALPHA_CUTOUT_THRESHOLD_DEFAULT = 0.1F;
     private static final float ALPHA_CUTOUT_THRESHOLD_CUTOUT_TERRAIN = 0.5F;
-    public static final RenderPipeline.Snippet GLOBALS_SNIPPET = RenderPipeline.builder().withBindGroupLayout(BindGroupLayouts.GLOBALS).buildSnippet();
-    public static final RenderPipeline.Snippet MATRICES_FOG_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
+    private static final RenderPipeline.Snippet GLOBALS_SNIPPET = RenderPipeline.builder().withBindGroupLayout(BindGroupLayouts.GLOBALS).buildSnippet();
+    private static final RenderPipeline.Snippet MATRICES_FOG_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
         .withBindGroupLayout(BindGroupLayouts.FOG)
         .buildSnippet();
-    public static final RenderPipeline.Snippet MATRICES_FOG_LIGHT_DIR_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
+    private static final RenderPipeline.Snippet MATRICES_FOG_LIGHT_DIR_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.LIGHTING)
         .buildSnippet();
     public static final RenderPipeline.Snippet OIT_OPAQUE_PARTS_THRESHOLD_SNIPPET = RenderPipeline.builder()
@@ -68,7 +_,7 @@
         .withColorTargetState(new ColorTargetState(Optional.of(BlendFunction.ADDITIVE), GpuFormat.RGBA16_FLOAT, 15))
         .withDepthStencilState(new DepthStencilState(CompareOp.GREATER_THAN_OR_EQUAL, false))
         .buildSnippet();
-    public static final RenderPipeline.Snippet GENERIC_BLOCKS_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
+    private static final RenderPipeline.Snippet GENERIC_BLOCKS_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.FOG)
         .withBindGroupLayout(BindGroupLayouts.SAMPLER0)
         .withVertexBinding(0, DefaultVertexFormat.BLOCK)
@@ -78,13 +_,13 @@
     private static final RenderPipeline.Snippet LIT_BLOCKS_SNIPPET = RenderPipeline.builder(GENERIC_BLOCKS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.SAMPLER2)
         .buildSnippet();
-    public static final RenderPipeline.Snippet TERRAIN_SNIPPET = RenderPipeline.builder(LIT_BLOCKS_SNIPPET)
+    private static final RenderPipeline.Snippet TERRAIN_SNIPPET = RenderPipeline.builder(LIT_BLOCKS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.CHUNK_SECTION)
         .withVertexShader("core/terrain")
         .withFragmentShader("core/terrain")
         .buildSnippet();
-    public static final RenderPipeline.Snippet BLOCK_SNIPPET = RenderPipeline.builder(LIT_BLOCKS_SNIPPET)
+    private static final RenderPipeline.Snippet BLOCK_SNIPPET = RenderPipeline.builder(LIT_BLOCKS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
         .withVertexShader("core/block")
@@ -121,7 +_,7 @@
         .withPrimitiveTopology(PrimitiveTopology.QUADS)
         .withDepthStencilState(DepthStencilState.DEFAULT)
         .buildSnippet();
-    public static final RenderPipeline.Snippet ENTITY_SNIPPET = RenderPipeline.builder(ENTITY_NO_LIGHTMAP_SNIPPET)
+    private static final RenderPipeline.Snippet ENTITY_SNIPPET = RenderPipeline.builder(ENTITY_NO_LIGHTMAP_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.LIGHTING)
         .withBindGroupLayout(BindGroupLayouts.SAMPLER2)
         .buildSnippet();
@@ -137,7 +_,7 @@
         .withDepthStencilState(new DepthStencilState(CompareOp.GREATER_THAN_OR_EQUAL, false))
         .withShaderDefine("ALPHA_CUTOUT", 0.1F)
         .buildSnippet();
-    public static final RenderPipeline.Snippet ENTITY_EMISSIVE_SNIPPET = RenderPipeline.builder(MATRICES_FOG_LIGHT_DIR_SNIPPET)
+    private static final RenderPipeline.Snippet ENTITY_EMISSIVE_SNIPPET = RenderPipeline.builder(MATRICES_FOG_LIGHT_DIR_SNIPPET)
         .withVertexShader("core/entity")
         .withFragmentShader("core/entity")
         .withBindGroupLayout(BindGroupLayouts.SAMPLER0)
@@ -151,7 +_,7 @@
         .withShaderDefine("NO_OVERLAY")
         .withShaderDefine("NO_CARDINAL_LIGHTING")
         .buildSnippet();
-    public static final RenderPipeline.Snippet BEACON_BEAM_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
+    private static final RenderPipeline.Snippet BEACON_BEAM_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
         .withVertexShader("core/rendertype_beacon_beam")
         .withFragmentShader("core/rendertype_beacon_beam")
         .withBindGroupLayout(BindGroupLayouts.SAMPLER0)
@@ -159,7 +_,7 @@
         .withPrimitiveTopology(PrimitiveTopology.QUADS)
         .withDepthStencilState(DepthStencilState.DEFAULT)
         .buildSnippet();
-    public static final RenderPipeline.Snippet ITEM_SNIPPET = RenderPipeline.builder(MATRICES_FOG_LIGHT_DIR_SNIPPET)
+    private static final RenderPipeline.Snippet ITEM_SNIPPET = RenderPipeline.builder(MATRICES_FOG_LIGHT_DIR_SNIPPET)
         .withVertexShader("core/item")
         .withFragmentShader("core/item")
         .withBindGroupLayout(BindGroupLayouts.SAMPLER0_SAMPLER1_SAMPLER2)
@@ -179,7 +_,7 @@
         .withDepthStencilState(new DepthStencilState(CompareOp.GREATER_THAN_OR_EQUAL, false))
         .withShaderDefine("ALPHA_CUTOUT", 0.1F)
         .buildSnippet();
-    public static final RenderPipeline.Snippet TEXT_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
+    private static final RenderPipeline.Snippet TEXT_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
         .withBindGroupLayout(BindGroupLayouts.SAMPLER0)
@@ -196,13 +_,13 @@
         .withFragmentShader("core/text_background")
         .withPrimitiveTopology(PrimitiveTopology.QUADS)
         .buildSnippet();
-    public static final RenderPipeline.Snippet WORLD_TEXT_SNIPPET = RenderPipeline.builder(TEXT_SNIPPET)
+    private static final RenderPipeline.Snippet WORLD_TEXT_SNIPPET = RenderPipeline.builder(TEXT_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.FOG)
         .withBindGroupLayout(BindGroupLayouts.SAMPLER2)
         .withVertexBinding(0, DefaultVertexFormat.POSITION_TEX_LIGHTMAP_COLOR)
         .withPrimitiveTopology(PrimitiveTopology.QUADS)
         .buildSnippet();
-    public static final RenderPipeline.Snippet END_PORTAL_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
+    private static final RenderPipeline.Snippet END_PORTAL_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
         .withBindGroupLayout(BindGroupLayouts.FOG)
@@ -214,7 +_,7 @@
         .withColorTargetState(ColorTargetState.DEFAULT)
         .withDepthStencilState(DepthStencilState.DEFAULT)
         .buildSnippet();
-    public static final RenderPipeline.Snippet CLOUDS_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
+    private static final RenderPipeline.Snippet CLOUDS_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
         .withVertexShader("core/clouds")
         .withFragmentShader("core/clouds")
         .withColorTargetState(new ColorTargetState(BlendFunction.TRANSLUCENT))
@@ -230,7 +_,7 @@
         .withBindGroupLayout(BindGroupLayouts.FOG)
         .withBindGroupLayout(BindGroupLayouts.CLOUD_INFO)
         .buildSnippet();
-    public static final RenderPipeline.Snippet LINES_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
+    private static final RenderPipeline.Snippet LINES_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
         .withVertexShader("core/rendertype_lines")
         .withFragmentShader("core/rendertype_lines")
         .withCull(false)
@@ -247,7 +_,7 @@
         .withVertexBinding(0, DefaultVertexFormat.POSITION_COLOR_NORMAL_LINE_WIDTH)
         .withPrimitiveTopology(PrimitiveTopology.LINES)
         .buildSnippet();
-    public static final RenderPipeline.Snippet DEBUG_FILLED_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
+    private static final RenderPipeline.Snippet DEBUG_FILLED_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
         .withVertexShader("core/position_color")
@@ -274,7 +_,7 @@
         .withPrimitiveTopology(PrimitiveTopology.POINTS)
         .withCull(false)
         .buildSnippet();
-    public static final RenderPipeline.Snippet PARTICLE_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
+    private static final RenderPipeline.Snippet PARTICLE_SNIPPET = RenderPipeline.builder(MATRICES_FOG_SNIPPET)
         .withVertexShader("core/particle")
         .withFragmentShader("core/particle")
         .withBindGroupLayout(BindGroupLayouts.SAMPLER0_SAMPLER2)
@@ -292,7 +_,7 @@
         .withPrimitiveTopology(PrimitiveTopology.QUADS)
         .withDepthStencilState(new DepthStencilState(CompareOp.GREATER_THAN_OR_EQUAL, false))
         .buildSnippet();
-    public static final RenderPipeline.Snippet WEATHER_SNIPPET = RenderPipeline.builder(PARTICLE_SNIPPET)
+    private static final RenderPipeline.Snippet WEATHER_SNIPPET = RenderPipeline.builder(PARTICLE_SNIPPET)
         .withColorTargetState(new ColorTargetState(BlendFunction.TRANSLUCENT))
         .withCull(false)
         .buildSnippet();
@@ -306,7 +_,7 @@
         .withPrimitiveTopology(PrimitiveTopology.QUADS)
         .withDepthStencilState(new DepthStencilState(CompareOp.GREATER_THAN_OR_EQUAL, false, 1.0F, 10.0F))
         .buildSnippet();
-    public static final RenderPipeline.Snippet GUI_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
+    private static final RenderPipeline.Snippet GUI_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
         .withVertexShader("core/gui")
@@ -315,7 +_,7 @@
         .withVertexBinding(0, DefaultVertexFormat.POSITION_COLOR)
         .withPrimitiveTopology(PrimitiveTopology.QUADS)
         .buildSnippet();
-    public static final RenderPipeline.Snippet GUI_TEXTURED_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
+    private static final RenderPipeline.Snippet GUI_TEXTURED_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
         .withVertexShader("core/position_tex_color")
@@ -325,12 +_,12 @@
         .withVertexBinding(0, DefaultVertexFormat.POSITION_TEX_COLOR)
         .withPrimitiveTopology(PrimitiveTopology.QUADS)
         .buildSnippet();
-    public static final RenderPipeline.Snippet GUI_TEXT_SNIPPET = RenderPipeline.builder(TEXT_SNIPPET)
+    private static final RenderPipeline.Snippet GUI_TEXT_SNIPPET = RenderPipeline.builder(TEXT_SNIPPET)
         .withShaderDefine("IS_GUI")
         .withColorTargetState(new ColorTargetState(BlendFunction.TRANSLUCENT))
         .withDepthStencilState(Optional.empty())
         .buildSnippet();
-    public static final RenderPipeline.Snippet OUTLINE_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
+    private static final RenderPipeline.Snippet OUTLINE_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
         .withVertexShader("core/rendertype_outline")
