--- a/net/minecraft/client/renderer/RenderPipelines.java
+++ b/net/minecraft/client/renderer/RenderPipelines.java
@@ -22,7 +_,6 @@
 @OnlyIn(Dist.CLIENT)
 public class RenderPipelines {
     private static final Map<Identifier, RenderPipeline> PIPELINES_BY_LOCATION = new HashMap<>();
-    private static final Map<Identifier, RenderPipeline> OPTIONAL_PIPELINES_BY_LOCATION = new HashMap<>();
     private static final float ALPHA_CUTOUT_THRESHOLD_DEFAULT = 0.1F;
     private static final float ALPHA_CUTOUT_THRESHOLD_CUTOUT_TERRAIN = 0.5F;
     private static final RenderPipeline.Snippet GLOBALS_SNIPPET = RenderPipeline.builder().withBindGroupLayout(BindGroupLayouts.GLOBALS).buildSnippet();
@@ -36,9 +_,9 @@
         .buildSnippet();
     private static final RenderPipeline.Snippet OIT_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
         .withShaderDefine("OIT")
-        .withShaderDefine("OIT_WAVELET_RANK", 2)
-        .withShaderDefine("OIT_COEFF_COUNT", LevelRenderer.OIT_COEFFICIENT_COUNT)
-        .withShaderDefine("OIT_COEFF_ATTACHMENT_COUNT", LevelRenderer.OIT_TRANSMITTANCE_TARGET_COUNT)
+        .withShaderDefine("WAVELET_RANK", 2)
+        .withShaderDefine("COEFF_COUNT", LevelRenderer.OIT_COEFFICIENT_COUNT)
+        .withShaderDefine("COEFF_ATTACHMENT_COUNT", LevelRenderer.OIT_TRANSMITTANCE_TARGET_COUNT)
         .buildSnippet();
     public static final RenderPipeline.Snippet OIT_DEPTH_BOUNDS_SNIPPET = RenderPipeline.builder(OIT_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.GLOBALS)
@@ -79,17 +_,8 @@
     private static final RenderPipeline.Snippet TERRAIN_SNIPPET = RenderPipeline.builder(LIT_BLOCKS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
         .withBindGroupLayout(BindGroupLayouts.CHUNK_SECTION)
-        .withBindGroupLayout(BindGroupLayouts.TERRAIN_INFO)
-        .withVertexShader("core/terrain")
-        .withFragmentShader("core/terrain")
-        .buildSnippet();
-    private static final RenderPipeline.Snippet MULTIDRAW_TERRAIN_SNIPPET = RenderPipeline.builder(LIT_BLOCKS_SNIPPET)
-        .withBindGroupLayout(BindGroupLayouts.PROJECTION)
-        .withBindGroupLayout(BindGroupLayouts.TERRAIN_INFO)
-        .withVertexBinding(1, DefaultVertexFormat.CHUNK_DATA_INSTANCED)
-        .withVertexShader("core/terrain")
-        .withFragmentShader("core/terrain")
-        .withShaderDefine("MULTIDRAW_TERRAIN")
+        .withVertexShader("core/terrain")
+        .withFragmentShader("core/terrain")
         .buildSnippet();
     private static final RenderPipeline.Snippet BLOCK_SNIPPET = RenderPipeline.builder(LIT_BLOCKS_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.PROJECTION)
@@ -196,6 +_,13 @@
         .withVertexBinding(0, DefaultVertexFormat.POSITION_TEX_COLOR)
         .withPrimitiveTopology(PrimitiveTopology.QUADS)
         .buildSnippet();
+    private static final RenderPipeline.Snippet TEXT_BACKGROUND_SNIPPET = RenderPipeline.builder(GLOBALS_SNIPPET)
+        .withBindGroupLayout(BindGroupLayouts.PROJECTION)
+        .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
+        .withVertexShader("core/text_background")
+        .withFragmentShader("core/text_background")
+        .withPrimitiveTopology(PrimitiveTopology.QUADS)
+        .buildSnippet();
     private static final RenderPipeline.Snippet WORLD_TEXT_SNIPPET = RenderPipeline.builder(TEXT_SNIPPET)
         .withBindGroupLayout(BindGroupLayouts.FOG)
         .withBindGroupLayout(BindGroupLayouts.SAMPLER2)
@@ -349,26 +_,13 @@
     public static final RenderPipeline SOLID_TERRAIN = register(
         RenderPipeline.builder(TERRAIN_SNIPPET).withLocation("pipeline/solid_terrain").withColorTargetState(ColorTargetState.DEFAULT).build()
     );
-    public static final RenderPipeline SOLID_TERRAIN_MULTIDRAW = register(
-        RenderPipeline.builder(MULTIDRAW_TERRAIN_SNIPPET)
-            .withLocation("pipeline/solid_terrain_multidraw")
-            .withColorTargetState(ColorTargetState.DEFAULT)
-            .build()
-    );
-    public static final RenderPipeline WIREFRAME = registerOptional(
+    public static final RenderPipeline WIREFRAME = register(
         RenderPipeline.builder(TERRAIN_SNIPPET)
             .withLocation("pipeline/wireframe")
             .withPolygonMode(PolygonMode.WIREFRAME)
             .withColorTargetState(ColorTargetState.DEFAULT)
             .build()
     );
-    public static final RenderPipeline WIREFRAME_MULTIDRAW = registerOptional(
-        RenderPipeline.builder(MULTIDRAW_TERRAIN_SNIPPET)
-            .withLocation("pipeline/wireframe_multidraw")
-            .withPolygonMode(PolygonMode.WIREFRAME)
-            .withColorTargetState(ColorTargetState.DEFAULT)
-            .build()
-    );
     public static final RenderPipeline CUTOUT_BLOCK = register(
         RenderPipeline.builder(BLOCK_SNIPPET)
             .withLocation("pipeline/cutout_block")
@@ -383,13 +_,6 @@
             .withColorTargetState(ColorTargetState.DEFAULT)
             .build()
     );
-    public static final RenderPipeline CUTOUT_TERRAIN_MULTIDRAW = register(
-        RenderPipeline.builder(MULTIDRAW_TERRAIN_SNIPPET)
-            .withLocation("pipeline/cutout_terrain_multidraw")
-            .withShaderDefine("ALPHA_CUTOUT", 0.5F)
-            .withColorTargetState(ColorTargetState.DEFAULT)
-            .build()
-    );
     public static final RenderPipeline TRANSLUCENT_TERRAIN = register(
         RenderPipeline.builder(TERRAIN_SNIPPET)
             .withLocation("pipeline/translucent_terrain")
@@ -397,13 +_,6 @@
             .withShaderDefine("ALPHA_CUTOUT", 0.1F)
             .build()
     );
-    public static final RenderPipeline TRANSLUCENT_TERRAIN_MULTIDRAW = register(
-        RenderPipeline.builder(MULTIDRAW_TERRAIN_SNIPPET)
-            .withLocation("pipeline/translucent_terrain_multidraw")
-            .withColorTargetState(new ColorTargetState(BlendFunction.TRANSLUCENT))
-            .withShaderDefine("ALPHA_CUTOUT", 0.1F)
-            .build()
-    );
     public static final OitPipelineSet OIT_TERRAIN = register(
         OitPipelineSet.builder(
                 "terrain",
@@ -411,22 +_,7 @@
                     .withVertexShader("core/terrain")
                     .withFragmentShader("core/terrain")
                     .withBindGroupLayout(BindGroupLayouts.CHUNK_SECTION)
-                    .withBindGroupLayout(BindGroupLayouts.TERRAIN_INFO)
-                    .withShaderDefine("ALPHA_CUTOUT", 0.1F)
-            )
-            .withAccumulateModifier(accumulate -> accumulate.withBindGroupLayout(BindGroupLayouts.SAMPLER2))
-            .build()
-    );
-    public static final OitPipelineSet OIT_TERRAIN_MULTIDRAW = register(
-        OitPipelineSet.builder(
-                "terrain_multidraw",
-                RenderPipeline.builder(GENERIC_BLOCKS_SNIPPET)
-                    .withVertexShader("core/terrain")
-                    .withFragmentShader("core/terrain")
-                    .withBindGroupLayout(BindGroupLayouts.TERRAIN_INFO)
-                    .withVertexBinding(1, DefaultVertexFormat.CHUNK_DATA_INSTANCED)
-                    .withShaderDefine("ALPHA_CUTOUT", 0.1F)
-                    .withShaderDefine("MULTIDRAW_TERRAIN")
+                    .withShaderDefine("ALPHA_CUTOUT", 0.1F)
             )
             .withAccumulateModifier(accumulate -> accumulate.withBindGroupLayout(BindGroupLayouts.SAMPLER2))
             .build()
@@ -442,13 +_,12 @@
     public static final OitPipelineSet OIT_TRANSLUCENT_BLOCK = register(
         OitPipelineSet.builder(
                 "translucent_block",
-                RenderPipeline.builder(GENERIC_BLOCKS_SNIPPET)
+                RenderPipeline.builder(LIT_BLOCKS_SNIPPET)
                     .withBindGroupLayout(BindGroupLayouts.DYNAMIC_TRANSFORMS)
                     .withVertexShader("core/block")
                     .withFragmentShader("core/block")
                     .withShaderDefine("ALPHA_CUTOUT", 0.1F)
             )
-            .withAccumulateModifier(accumulate -> accumulate.withBindGroupLayout(BindGroupLayouts.SAMPLER2))
             .build()
     );
     public static final RenderPipeline ARMOR_CUTOUT_NO_CULL = register(
@@ -746,20 +_,21 @@
             .build()
     );
     public static final OitPipelineSet OIT_ITEM_GLINT = register(
-        OitPipelineSet.builder("item_glint", RenderPipeline.builder(OIT_ITEM_SNIPPET))
+        OitPipelineSet.builder("item_glint", RenderPipeline.builder(OIT_ITEM_SNIPPET, GLINT_SNIPPET))
             .withAccumulateModifier(
-                accumulate -> accumulate.withSnippet(GLINT_SNIPPET)
-                    .withBindGroupLayout(BindGroupLayouts.GLOBALS)
+                accumulate -> accumulate.withBindGroupLayout(BindGroupLayouts.GLOBALS)
                     .withBindGroupLayout(BindGroupLayouts.SAMPLER1)
                     .withBindGroupLayout(BindGroupLayouts.SAMPLER2)
             )
             .build()
     );
     public static final OitPipelineSet OIT_ITEM_GLINT_SPECIAL = register(
-        OitPipelineSet.builder("item_glint_special", RenderPipeline.builder(OIT_ITEM_SNIPPET).withVertexBinding(0, DefaultVertexFormat.ENTITY_GLINT_SPECIAL))
+        OitPipelineSet.builder(
+                "item_glint_special",
+                RenderPipeline.builder(OIT_ITEM_SNIPPET, GLINT_SPECIAL_SNIPPET).withVertexBinding(0, DefaultVertexFormat.ENTITY_GLINT_SPECIAL)
+            )
             .withAccumulateModifier(
-                accumulate -> accumulate.withSnippet(GLINT_SPECIAL_SNIPPET)
-                    .withBindGroupLayout(BindGroupLayouts.GLOBALS)
+                accumulate -> accumulate.withBindGroupLayout(BindGroupLayouts.GLOBALS)
                     .withBindGroupLayout(BindGroupLayouts.SAMPLER1)
                     .withBindGroupLayout(BindGroupLayouts.SAMPLER2)
             )
@@ -825,6 +_,26 @@
     );
     public static final OitPipelineSet OIT_TEXT = register(OitPipelineSet.builder("text", RenderPipeline.builder(WORLD_TEXT_SNIPPET)).build());
     public static final RenderPipeline GUI_TEXT = register(RenderPipeline.builder(GUI_TEXT_SNIPPET).withLocation("pipeline/gui_text").build());
+    public static final RenderPipeline TEXT_BACKGROUND = register(
+        RenderPipeline.builder(TEXT_BACKGROUND_SNIPPET)
+            .withBindGroupLayout(BindGroupLayouts.FOG)
+            .withLocation("pipeline/text_background")
+            .withColorTargetState(new ColorTargetState(BlendFunction.TRANSLUCENT))
+            .withDepthStencilState(DepthStencilState.DEFAULT)
+            .withBindGroupLayout(BindGroupLayouts.SAMPLER2)
+            .withVertexBinding(0, DefaultVertexFormat.POSITION_COLOR_LIGHTMAP)
+            .build()
+    );
+    public static final OitPipelineSet OIT_TEXT_BACKGROUND = register(
+        OitPipelineSet.builder(
+                "text_background",
+                RenderPipeline.builder(TEXT_BACKGROUND_SNIPPET)
+                    .withBindGroupLayout(BindGroupLayouts.FOG)
+                    .withBindGroupLayout(BindGroupLayouts.SAMPLER2)
+                    .withVertexBinding(0, DefaultVertexFormat.POSITION_COLOR_LIGHTMAP)
+            )
+            .build()
+    );
     public static final RenderPipeline TEXT_GRAYSCALE = register(
         RenderPipeline.builder(WORLD_TEXT_SNIPPET)
             .withLocation("pipeline/text_grayscale")
@@ -867,6 +_,33 @@
             .withDepthStencilState(Optional.empty())
             .build()
     );
+    public static final OitPipelineSet OIT_TEXT_SEE_THROUGH = register(
+        OitPipelineSet.builder(
+                "text_see_through", RenderPipeline.builder(TEXT_SNIPPET).withShaderDefine("IS_SEE_THROUGH").withShaderDefine("OIT_FORCE_ZERO_DEPTH")
+            )
+            .withoutDepthTest()
+            .build()
+    );
+    public static final RenderPipeline TEXT_BACKGROUND_SEE_THROUGH = register(
+        RenderPipeline.builder(TEXT_BACKGROUND_SNIPPET)
+            .withLocation("pipeline/text_background_see_through")
+            .withVertexBinding(0, DefaultVertexFormat.POSITION_COLOR)
+            .withShaderDefine("IS_SEE_THROUGH")
+            .withColorTargetState(new ColorTargetState(BlendFunction.TRANSLUCENT))
+            .withDepthStencilState(Optional.empty())
+            .build()
+    );
+    public static final OitPipelineSet OIT_TEXT_BACKGROUND_SEE_THROUGH = register(
+        OitPipelineSet.builder(
+                "text_background_see_through",
+                RenderPipeline.builder(TEXT_BACKGROUND_SNIPPET)
+                    .withVertexBinding(0, DefaultVertexFormat.POSITION_COLOR)
+                    .withShaderDefine("IS_SEE_THROUGH")
+                    .withShaderDefine("OIT_FORCE_ZERO_DEPTH")
+            )
+            .withoutDepthTest()
+            .build()
+    );
     public static final RenderPipeline TEXT_GRAYSCALE_SEE_THROUGH = register(
         RenderPipeline.builder(TEXT_SNIPPET)
             .withLocation("pipeline/text_grayscale_see_through")
@@ -876,6 +_,17 @@
             .withDepthStencilState(Optional.empty())
             .build()
     );
+    public static final OitPipelineSet OIT_TEXT_GRAYSCALE_SEE_THROUGH = register(
+        OitPipelineSet.builder(
+                "text_grayscale_see_through",
+                RenderPipeline.builder(TEXT_SNIPPET)
+                    .withShaderDefine("IS_GRAYSCALE")
+                    .withShaderDefine("IS_SEE_THROUGH")
+                    .withShaderDefine("OIT_FORCE_ZERO_DEPTH")
+            )
+            .withoutDepthTest()
+            .build()
+    );
     public static final RenderPipeline LIGHTNING = register(
         RenderPipeline.builder(LIGHTNING_SNIPPET)
             .withLocation("pipeline/lightning")
@@ -919,12 +_,6 @@
         RenderPipeline.builder(LINES_SNIPPET)
             .withLocation("pipeline/lines_translucent")
             .withColorTargetState(new ColorTargetState(BlendFunction.TRANSLUCENT))
-            .build()
-    );
-    public static final RenderPipeline LINES_TRANSLUCENT_NO_DEPTH_WRITE = register(
-        RenderPipeline.builder(LINES_SNIPPET)
-            .withLocation("pipeline/lines_translucent_no_depth_write")
-            .withColorTargetState(new ColorTargetState(BlendFunction.TRANSLUCENT))
             .withDepthStencilState(new DepthStencilState(CompareOp.GREATER_THAN_OR_EQUAL, false))
             .build()
     );
@@ -941,9 +_,12 @@
     public static final RenderPipeline SECONDARY_BLOCK_OUTLINE = register(
         RenderPipeline.builder(LINES_SNIPPET)
             .withLocation("pipeline/secondary_block_outline")
-            .withColorTargetState(ColorTargetState.DEFAULT)
-            .withDepthStencilState(DepthStencilState.DEFAULT)
+            .withColorTargetState(new ColorTargetState(BlendFunction.TRANSLUCENT))
+            .withDepthStencilState(new DepthStencilState(CompareOp.GREATER_THAN_OR_EQUAL, false))
             .build()
+    );
+    public static final OitPipelineSet OIT_SECONDARY_BLOCK_OUTLINE = register(
+        OitPipelineSet.builder("secondary_block_outline", RenderPipeline.builder(OIT_LINES_SNIPPET)).build()
     );
     public static final RenderPipeline DEBUG_POINTS = register(
         RenderPipeline.builder(DEBUG_POINTS_SNIPPET)
@@ -1214,18 +_,8 @@
             .withBindGroupLayout(BindGroupLayouts.CURRENT_SPRITE_NEXT_SPRITE)
             .build()
     );
-    public static final RenderPipeline BLIT_DEPTH_BOUNDS = register(
-        RenderPipeline.builder(GLOBALS_SNIPPET)
-            .withLocation("pipeline/blit_depth_bounds")
-            .withVertexShader("core/screenquad")
-            .withFragmentShader("core/blit_screen")
-            .withBindGroupLayout(BindGroupLayouts.IN_SAMPLER)
-            .withPrimitiveTopology(PrimitiveTopology.TRIANGLES)
-            .withColorTargetState(new ColorTargetState(Optional.empty(), GpuFormat.RGBA32_FLOAT, 15))
-            .build()
-    );
     public static final RenderPipeline OIT_DEPTH_BOUNDS_CULL = register(
-        RenderPipeline.builder(OIT_SNIPPET)
+        RenderPipeline.builder()
             .withVertexShader("core/screenquad")
             .withFragmentShader("core/oit_depth_bounds_cull")
             .withLocation("pipeline/oit_depth_bounds_cull")
@@ -1236,27 +_,6 @@
             .withDepthStencilState(DepthStencilState.DEFAULT)
             .build()
     );
-    public static final RenderPipeline BLIT_DEPTH_DURING_DEPTH_BOUNDS = register(
-        RenderPipeline.builder(GLOBALS_SNIPPET)
-            .withLocation("pipeline/blit_depth_during_depth_bounds")
-            .withVertexShader("core/screenquad")
-            .withFragmentShader("core/blit_depth")
-            .withBindGroupLayout(BindGroupLayouts.IN_SAMPLER)
-            .withPrimitiveTopology(PrimitiveTopology.TRIANGLES)
-            .withDepthStencilState(new DepthStencilState(CompareOp.ALWAYS_PASS, true))
-            .withColorTargetState(new ColorTargetState(Optional.empty(), GpuFormat.RGBA32_FLOAT, 0))
-            .build()
-    );
-    public static final RenderPipeline BLIT_DEPTH = register(
-        RenderPipeline.builder(GLOBALS_SNIPPET)
-            .withLocation("pipeline/blit_depth")
-            .withVertexShader("core/screenquad")
-            .withFragmentShader("core/blit_depth")
-            .withBindGroupLayout(BindGroupLayouts.IN_SAMPLER)
-            .withPrimitiveTopology(PrimitiveTopology.TRIANGLES)
-            .withDepthStencilState(new DepthStencilState(CompareOp.ALWAYS_PASS, true))
-            .build()
-    );
     public static final RenderPipeline OIT_COMPOSITE = register(
         RenderPipeline.builder(OIT_SNIPPET)
             .withBindGroupLayout(BindGroupLayouts.PROJECTION)
@@ -1292,16 +_,7 @@
         return oitPipelineSet;
     }
 
-    private static RenderPipeline registerOptional(RenderPipeline pipeline) {
-        OPTIONAL_PIPELINES_BY_LOCATION.put(pipeline.getLocation(), pipeline);
-        return pipeline;
-    }
-
-    public static List<RenderPipeline> requiredPipelines() {
+    public static List<RenderPipeline> getStaticPipelines() {
         return PIPELINES_BY_LOCATION.values().stream().toList();
-    }
-
-    public static List<RenderPipeline> optionalPipelines() {
-        return OPTIONAL_PIPELINES_BY_LOCATION.values().stream().toList();
     }
 }
