--- a/net/minecraft/client/renderer/SectionOcclusionGraph.java
+++ b/net/minecraft/client/renderer/SectionOcclusionGraph.java
@@ -275,8 +_,8 @@
                     if (storage.sectionTree.add(node.section)) {
                         onSectionAdded.accept(node.section);
                     }
-                } else if (node.section.sectionMesh.compareAndSet(CompiledSectionMesh.UNCOMPILED, CompiledSectionMesh.EMPTY)) {
-                    node.section.updateUploadTime();
+                } else {
+                    node.section.sectionMesh.compareAndSet(CompiledSectionMesh.UNCOMPILED, CompiledSectionMesh.EMPTY);
                 }
 
                 boolean distantFromCamera = Math.abs(SectionPos.x(sectionNode) - cameraSectionPos.x()) > MINIMUM_ADVANCED_CULLING_SECTION_DISTANCE
@@ -397,6 +_,7 @@
                 SectionRenderDispatcher.RenderSection section = this.viewArea.getRenderSection(sectionNode);
                 if (section != null) {
                     this.schedulePropagationFrom(section);
+                    section.setWasPreviouslyEmpty(true);
                 }
             }
         }
