--- a/net/minecraft/client/renderer/SubmitNodeStorage.java
+++ b/net/minecraft/client/renderer/SubmitNodeStorage.java
@@ -11,7 +_,6 @@
 import net.minecraft.client.renderer.entity.state.EntityRenderState;
 import net.minecraft.client.renderer.feature.ModelFeatureRenderer;
 import net.minecraft.client.renderer.feature.phase.FeatureRenderPhase;
-import net.minecraft.client.renderer.feature.phase.TranslucentFeatureRenderPhase;
 import net.minecraft.client.renderer.gizmos.DrawableGizmoPrimitives;
 import net.minecraft.client.renderer.item.ItemStackRenderState;
 import net.minecraft.client.renderer.rendertype.RenderType;
@@ -32,7 +_,6 @@
 @OnlyIn(Dist.CLIENT)
 public class SubmitNodeStorage implements SubmitNodeCollector {
     private final Int2ObjectAVLTreeMap<SubmitNodeCollection> submitsPerOrder = new Int2ObjectAVLTreeMap<>();
-    private final TranslucentFeatureRenderPhase seeThrough = new TranslucentFeatureRenderPhase();
     private boolean useImprovedTransparency;
 
     public void setUseImprovedTransparency(boolean useImprovedTransparency) {
@@ -51,11 +_,7 @@
     }
 
     public SubmitNodeCollection order(int order) {
-        return this.submitsPerOrder.computeIfAbsent(order, var1 -> new SubmitNodeCollection(this.useImprovedTransparency, this.seeThrough));
-    }
-
-    public TranslucentFeatureRenderPhase seeThrough() {
-        return this.seeThrough;
+        return this.submitsPerOrder.computeIfAbsent(order, var1 -> new SubmitNodeCollection(this.useImprovedTransparency));
     }
 
     @Override
@@ -87,11 +_,6 @@
     }
 
     @Override
-    public void submitTextBackground(PoseStack poseStack, float x0, float y0, float x1, float y1, int color, Font.DisplayMode displayMode, int lightCoords) {
-        this.order(0).submitTextBackground(poseStack, x0, y0, x1, y1, color, displayMode, lightCoords);
-    }
-
-    @Override
     public void submitFlame(PoseStack poseStack, EntityRenderState renderState, Quaternionf rotation) {
         this.order(0).submitFlame(poseStack, renderState, rotation);
     }
@@ -204,8 +_,5 @@
 
             return empty;
         });
-        if (!this.seeThrough.isEmpty()) {
-            consumer.accept(this.seeThrough);
-        }
     }
 }
