--- a/net/minecraft/client/renderer/SubmitNodeStorage.java
+++ b/net/minecraft/client/renderer/SubmitNodeStorage.java
@@ -28,10 +_,8 @@
 import org.joml.Quaternionf;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.extensions.SubmitNodeStorageExtension;
-
 @OnlyIn(Dist.CLIENT)
-public class SubmitNodeStorage implements SubmitNodeCollector, SubmitNodeStorageExtension {
+public class SubmitNodeStorage implements SubmitNodeCollector {
     private final Int2ObjectAVLTreeMap<SubmitNodeCollection> submitsPerOrder = new Int2ObjectAVLTreeMap<>();
     private boolean useImprovedTransparency;
 
