--- a/net/minecraft/client/renderer/UiLightmap.java
+++ b/net/minecraft/client/renderer/UiLightmap.java
@@ -8,11 +_,18 @@
 
 @OnlyIn(Dist.CLIENT)
 public class UiLightmap implements AutoCloseable {
-    private final DynamicTexture texture = new DynamicTexture("UI Lightmap", 1, 1, false);
+    // Neo: change the UI lightmap to match the level lightmap such that special item models and PiP renderers can use
+    // the same light coordinates in level and in UI rendering. (An OOB read will make the vertex transparent).
+    private static final int TEXTURE_SIZE = Lightmap.TEXTURE_SIZE;
+    private final DynamicTexture texture = new DynamicTexture("UI Lightmap", TEXTURE_SIZE, TEXTURE_SIZE, false);
 
     public UiLightmap() {
         NativeImage pixels = this.texture.getPixels();
-        pixels.setPixel(0, 0, -1);
+        for (int i = 0; i < TEXTURE_SIZE; ++i) {
+            for (int j = 0; j < TEXTURE_SIZE; ++j) {
+                pixels.setPixel(i, j, -1);
+            }
+        }
         this.texture.upload();
     }
 
