--- a/net/minecraft/client/renderer/WeatherEffectRenderer.java
+++ b/net/minecraft/client/renderer/WeatherEffectRenderer.java
@@ -46,8 +_,8 @@
     private final float[] columnSizeZ = new float[1024];
     private final GameRenderer gameRenderer;
     private final TextureManager textureManager;
-    private @Nullable AbstractTexture rainTexture;
-    private @Nullable AbstractTexture snowTexture;
+    private AbstractTexture rainTexture;
+    private AbstractTexture snowTexture;
     private @Nullable GpuBuffer vertexBuffer;
     private @Nullable PrimitiveTopology primitiveTopology;
     private int indexCount;
@@ -66,6 +_,8 @@
         Minecraft minecraft = Minecraft.getInstance();
         this.gameRenderer = minecraft.gameRenderer;
         this.textureManager = minecraft.getTextureManager();
+        this.rainTexture = this.textureManager.getTexture(RAIN_LOCATION);
+        this.snowTexture = this.textureManager.getTexture(SNOW_LOCATION);
     }
 
     public void extractRenderState(ClientLevel level, float partialTicks, Vec3 cameraPos, WeatherRenderState renderState) {
@@ -105,7 +_,7 @@
     }
 
     private void renderWeather(RenderPass renderPass, AbstractTexture texture, int startColumn, int columnCount) {
-        renderPass.setUniform("Sampler0", texture.getTextureView(), texture.getSampler());
+        renderPass.bindTexture("Sampler0", texture.getTextureView(), texture.getSampler());
         renderPass.drawIndexed(columnCount * 6, 1, startColumn * 6, 0, 0);
     }
 
@@ -165,7 +_,7 @@
             renderPass.setPipeline(RenderSystem.getCompiledPipeline(renderPipeline));
             RenderSystem.bindDefaultUniforms(renderPass);
             renderPass.setUniform("DynamicTransforms", dynamicTransforms);
-            renderPass.setUniform("Sampler2", this.gameRenderer.lightmap(), RenderSystem.getSamplerCache().getClampToEdge(FilterMode.LINEAR));
+            renderPass.bindTexture("Sampler2", this.gameRenderer.lightmap(), RenderSystem.getSamplerCache().getClampToEdge(FilterMode.LINEAR));
             renderPass.setIndexBuffer(indexBuffer, indexType);
             renderPass.setVertexBuffer(0, this.vertexBuffer.slice());
             this.renderWeather(renderPass, this.rainTexture, 0, renderState.rainColumns.size());
@@ -206,7 +_,7 @@
             for (WeatherEffectRenderer.ColumnInstance column : columns) {
                 float relativeX = (float)(column.x + 0.5 - cameraPos.x);
                 float relativeZ = (float)(column.z + 0.5 - cameraPos.z);
-                float distanceSq = Mth.lengthSquared(relativeX, relativeZ);
+                float distanceSq = (float)Mth.lengthSquared(relativeX, relativeZ);
                 float alpha = Mth.lerp(Math.min(distanceSq / radiusSq, 1.0F), maxAlpha, 0.5F) * intensity;
                 int color = ARGB.white(alpha);
                 int index = (column.z - Mth.floor(cameraPos.z) + 16) * 32 + column.x - Mth.floor(cameraPos.x) + 16;
