--- a/net/minecraft/client/renderer/WeatherEffectRenderer.java
+++ b/net/minecraft/client/renderer/WeatherEffectRenderer.java
@@ -92,7 +_,19 @@
         }
     }
 
+    /**
+     * @deprecated Neo: use {@link #render(MultiBufferSource, Vec3, WeatherRenderState, net.minecraft.client.renderer.state.LevelRenderState)} instead
+     */
+    @Deprecated
     public void render(MultiBufferSource bufferSource, Vec3 cameraPos, WeatherRenderState renderState) {
+        this.render(bufferSource, cameraPos, renderState, null);
+    }
+
+    public void render(MultiBufferSource bufferSource, Vec3 cameraPos, WeatherRenderState renderState, net.minecraft.client.renderer.state.@org.jspecify.annotations.Nullable LevelRenderState levelRenderState) {
+        if (levelRenderState != null && levelRenderState.customWeatherEffectRenderer != null && levelRenderState.customWeatherEffectRenderer.renderSnowAndRain(levelRenderState, renderState, bufferSource, cameraPos)) {
+            return;
+        }
+
         if (!renderState.rainColumns.isEmpty()) {
             RenderType renderType = RenderTypes.weather(RAIN_LOCATION, Minecraft.useShaderTransparency());
             this.renderInstances(bufferSource.getBuffer(renderType), renderState.rainColumns, cameraPos, 1.0F, renderState.radius, renderState.intensity);
@@ -158,6 +_,10 @@
     }
 
     public void tickRainParticles(ClientLevel level, Camera camera, int ticks, ParticleStatus particleStatus, int weatherRadius) {
+        var customWeatherEffectRenderer = net.neoforged.neoforge.client.CustomEnvironmentEffectsRendererManager.getCustomWeatherEffectRenderer(level, camera.position());
+        if (customWeatherEffectRenderer != null && customWeatherEffectRenderer.tickRain(level, ticks, camera))
+            return;
+
         float rainLevel = level.getRainLevel(1.0F);
         if (!(rainLevel <= 0.0F)) {
             RandomSource random = RandomSource.create(ticks * 312987231L);
