--- a/com/mojang/blaze3d/resource/RenderTargetDescriptor.java
+++ b/com/mojang/blaze3d/resource/RenderTargetDescriptor.java
@@ -20,13 +_,13 @@
     }
 
     public void prepare(RenderTarget resource) {
-        if (this.color != null && this.depth != null && this.color.clearColor != null && this.depth.clearColor != null) {
+        if (this.color != null && this.depth != null) {
             RenderSystem.getDevice()
                 .createCommandEncoder()
                 .clearColorAndDepthTextures(resource.getColorTexture(), this.color.clearColor, resource.getDepthTexture(), this.depth.clearColor.x());
-        } else if (this.color != null && this.color.clearColor != null) {
+        } else if (this.color != null) {
             RenderSystem.getDevice().createCommandEncoder().clearColorTexture(resource.getColorTexture(), this.color.clearColor);
-        } else if (this.depth != null && this.depth.clearColor != null) {
+        } else if (this.depth != null) {
             RenderSystem.getDevice().createCommandEncoder().clearDepthTexture(resource.getDepthTexture(), this.depth.clearColor.x());
         }
     }
@@ -45,7 +_,7 @@
                 && Objects.equals(this.depth, descriptor.depth);
     }
 
-    public record TextureProperties(@Nullable Vector4fc clearColor, GpuFormat format) {
+    public record TextureProperties(Vector4fc clearColor, GpuFormat format) {
         public static final RenderTargetDescriptor.TextureProperties DEFAULT_DEPTH = new RenderTargetDescriptor.TextureProperties(
             new Vector4f(0.0F, 0.0F, 0.0F, 0.0F), GpuFormat.D32_FLOAT
         );
