--- a/net/minecraft/world/level/storage/loot/functions/ApplyBonusCount.java
+++ b/net/minecraft/world/level/storage/loot/functions/ApplyBonusCount.java
@@ -4,8 +_,8 @@
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.List;
 import java.util.Map;
-import java.util.Optional;
 import java.util.Set;
 import java.util.function.Function;
 import java.util.stream.Collectors;
@@ -43,8 +_,8 @@
     private final Holder<Enchantment> enchantment;
     private final ApplyBonusCount.Formula formula;
 
-    private ApplyBonusCount(Optional<Holder<LootItemCondition>> condition, Holder<Enchantment> enchantment, ApplyBonusCount.Formula formula) {
-        super(condition);
+    private ApplyBonusCount(List<LootItemCondition> predicates, Holder<Enchantment> enchantment, ApplyBonusCount.Formula formula) {
+        super(predicates);
         this.enchantment = enchantment;
         this.formula = formula;
     }
@@ -64,6 +_,10 @@
         ItemInstance tool = context.getOptionalParameter(LootContextParams.TOOL);
         if (tool != null) {
             int level = EnchantmentHelper.getItemEnchantmentLevel(this.enchantment, tool);
+
+            // Neo: Allow mods to modify the enchantment level for loot calculations with relevant context.
+            level = net.neoforged.neoforge.event.EventHooks.getBlockLootEnchantmentLevel(tool, this.enchantment, level, context);
+
             int newCount = this.formula.calculateNewCount(context.getRandom(), itemStack.getCount(), level);
             itemStack.setCount(newCount);
         }
