--- a/net/minecraft/world/level/storage/loot/functions/EnchantWithLevelsFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/EnchantWithLevelsFunction.java
@@ -9,9 +_,9 @@
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderSet;
 import net.minecraft.core.RegistryAccess;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.tags.EnchantmentTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.context.ContextKey;
@@ -23,28 +_,28 @@
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 
 public class EnchantWithLevelsFunction extends LootItemConditionalFunction {
     public static final MapCodec<EnchantWithLevelsFunction> MAP_CODEC = RecordCodecBuilder.mapCodec(
         i -> commonFields(i)
             .and(
                 i.group(
-                    ContextIntProviders.CODEC.fieldOf("levels").forGetter(f -> f.levels),
-                    RegistryCodecs.holderSet(Registries.ENCHANTMENT).optionalFieldOf("options").forGetter(f -> f.options),
+                    NumberProviders.CODEC.fieldOf("levels").forGetter(f -> f.levels),
+                    RegistryCodecs.homogeneousList(Registries.ENCHANTMENT).optionalFieldOf("options").forGetter(f -> f.options),
                     Codec.BOOL.optionalFieldOf("include_additional_cost_component", false).forGetter(f -> f.includeAdditionalCostComponent)
                 )
             )
             .apply(i, EnchantWithLevelsFunction::new)
     );
-    private final Holder<ContextIntProvider> levels;
+    private final Holder<NumberProvider> levels;
     private final Optional<HolderSet<Enchantment>> options;
     private final boolean includeAdditionalCostComponent;
 
     private EnchantWithLevelsFunction(
         Optional<Holder<LootItemCondition>> condition,
-        Holder<ContextIntProvider> levels,
+        Holder<NumberProvider> levels,
         Optional<HolderSet<Enchantment>> options,
         boolean includeAdditionalCostComponent
     ) {
@@ -86,16 +_,16 @@
         return result;
     }
 
-    public static EnchantWithLevelsFunction.Builder enchantWithLevels(HolderGetter<Enchantment> enchantments, Holder<ContextIntProvider> levels) {
+    public static EnchantWithLevelsFunction.Builder enchantWithLevels(HolderGetter<Enchantment> enchantments, Holder<NumberProvider> levels) {
         return new EnchantWithLevelsFunction.Builder(levels).withOptions(enchantments.getOrThrow(EnchantmentTags.ON_RANDOM_LOOT));
     }
 
     public static class Builder extends LootItemConditionalFunction.Builder<EnchantWithLevelsFunction.Builder> {
-        private final Holder<ContextIntProvider> levels;
+        private final Holder<NumberProvider> levels;
         private Optional<HolderSet<Enchantment>> options = Optional.empty();
         private boolean includeAdditionalCostComponent = false;
 
-        public Builder(Holder<ContextIntProvider> levels) {
+        public Builder(Holder<NumberProvider> levels) {
             this.levels = levels;
         }
 
