--- a/net/minecraft/world/level/storage/loot/functions/EnchantedCountIncreaseFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/EnchantedCountIncreaseFunction.java
@@ -3,10 +_,11 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import java.util.Optional;
+import java.util.List;
 import java.util.Set;
 import net.minecraft.core.Holder;
-import net.minecraft.core.HolderGetter;
+import net.minecraft.core.HolderLookup;
+import net.minecraft.core.registries.Registries;
 import net.minecraft.util.context.ContextKey;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
@@ -19,8 +_,8 @@
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProvider;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProviders;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 
 public class EnchantedCountIncreaseFunction extends LootItemConditionalFunction {
     public static final int NO_LIMIT = 0;
@@ -29,20 +_,18 @@
             .and(
                 i.group(
                     Enchantment.CODEC.fieldOf("enchantment").forGetter(f -> f.enchantment),
-                    ContextFloatProviders.CODEC.fieldOf("count").forGetter(f -> f.count),
+                    NumberProviders.CODEC.fieldOf("count").forGetter(f -> f.count),
                     Codec.INT.optionalFieldOf("limit", 0).forGetter(f -> f.limit)
                 )
             )
             .apply(i, EnchantedCountIncreaseFunction::new)
     );
     private final Holder<Enchantment> enchantment;
-    private final Holder<ContextFloatProvider> count;
+    private final NumberProvider count;
     private final int limit;
 
-    private EnchantedCountIncreaseFunction(
-        Optional<Holder<LootItemCondition>> condition, Holder<Enchantment> enchantment, Holder<ContextFloatProvider> count, int limit
-    ) {
-        super(condition);
+    private EnchantedCountIncreaseFunction(List<LootItemCondition> predicates, Holder<Enchantment> enchantment, NumberProvider count, int limit) {
+        super(predicates);
         this.enchantment = enchantment;
         this.count = count;
         this.limit = limit;
@@ -61,7 +_,7 @@
     @Override
     public void validate(ValidationContext context) {
         super.validate(context);
-        Validatable.validateHolder(context, "count", this.count);
+        Validatable.validate(context, "count", this.count);
     }
 
     private boolean hasLimit() {
@@ -73,11 +_,19 @@
         Entity killer = context.getOptionalParameter(LootContextParams.ATTACKING_ENTITY);
         if (killer instanceof LivingEntity entity) {
             int level = EnchantmentHelper.getEnchantmentLevel(this.enchantment, entity);
+            // Purpur start - Add an option to fix MC-3304 projectile looting
+            if (org.purpurmc.purpur.PurpurConfig.fixProjectileLootingTransfer && context.getOptionalParameter(LootContextParams.DIRECT_ATTACKING_ENTITY) instanceof net.minecraft.world.entity.projectile.arrow.AbstractArrow arrow) {
+                level = arrow.actualEnchantments.getLevel(this.enchantment);
+            }
+            // Purpur end - Add an option to fix MC-3304 projectile looting
+            // Neo: Allow mods to modify the enchantment level for loot calculations with relevant context.
+            level = net.neoforged.neoforge.event.EventHooks.getEntityLootEnchantmentLevel(this.enchantment, level, context);
+
             if (level == 0) {
                 return itemStack;
             }
 
-            float addition = level * this.count.value().getFloat(context);
+            float addition = level * this.count.getFloat(context);
             itemStack.grow(Math.round(addition));
             if (this.hasLimit()) {
                 itemStack.limitSize(this.limit);
@@ -87,16 +_,17 @@
         return itemStack;
     }
 
-    public static EnchantedCountIncreaseFunction.Builder lootingMultiplier(HolderGetter<Enchantment> enchantments, Holder<ContextFloatProvider> count) {
+    public static EnchantedCountIncreaseFunction.Builder lootingMultiplier(HolderLookup.Provider registries, NumberProvider count) {
+        HolderLookup.RegistryLookup<Enchantment> enchantments = registries.lookupOrThrow(Registries.ENCHANTMENT);
         return new EnchantedCountIncreaseFunction.Builder(enchantments.getOrThrow(Enchantments.LOOTING), count);
     }
 
     public static class Builder extends LootItemConditionalFunction.Builder<EnchantedCountIncreaseFunction.Builder> {
         private final Holder<Enchantment> enchantment;
-        private final Holder<ContextFloatProvider> count;
+        private final NumberProvider count;
         private int limit = 0;
 
-        public Builder(Holder<Enchantment> enchantment, Holder<ContextFloatProvider> count) {
+        public Builder(Holder<Enchantment> enchantment, NumberProvider count) {
             this.enchantment = enchantment;
             this.count = count;
         }
@@ -112,7 +_,7 @@
 
         @Override
         public LootItemFunction build() {
-            return new EnchantedCountIncreaseFunction(this.getCondition(), this.enchantment, this.count, this.limit);
+            return new EnchantedCountIncreaseFunction(this.getConditions(), this.enchantment, this.count, this.limit);
         }
     }
 }
