--- a/net/minecraft/world/level/storage/loot/functions/LimitCount.java
+++ b/net/minecraft/world/level/storage/loot/functions/LimitCount.java
@@ -5,7 +_,7 @@
 import java.util.Optional;
 import net.minecraft.core.Holder;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.level.storage.loot.IntLimit;
+import net.minecraft.world.level.storage.loot.IntRange;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.Validatable;
 import net.minecraft.world.level.storage.loot.ValidationContext;
@@ -13,11 +_,11 @@
 
 public class LimitCount extends LootItemConditionalFunction {
     public static final MapCodec<LimitCount> MAP_CODEC = RecordCodecBuilder.mapCodec(
-        i -> commonFields(i).and(IntLimit.CODEC.fieldOf("limit").forGetter(f -> f.limit)).apply(i, LimitCount::new)
+        i -> commonFields(i).and(IntRange.CODEC.fieldOf("limit").forGetter(f -> f.limit)).apply(i, LimitCount::new)
     );
-    private final IntLimit limit;
+    private final IntRange limit;
 
-    private LimitCount(Optional<Holder<LootItemCondition>> condition, IntLimit limit) {
+    private LimitCount(Optional<Holder<LootItemCondition>> condition, IntRange limit) {
         super(condition);
         this.limit = limit;
     }
@@ -40,7 +_,7 @@
         return itemStack;
     }
 
-    public static LootItemConditionalFunction.Builder<?> limitCount(IntLimit limit) {
+    public static LootItemConditionalFunction.Builder<?> limitCount(IntRange limit) {
         return simpleBuilder(conditions -> new LimitCount(conditions, limit));
     }
 }
