--- a/net/minecraft/world/level/storage/loot/functions/SetAttributesFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetAttributesFunction.java
@@ -24,8 +_,8 @@
 import net.minecraft.world.level.storage.loot.Validatable;
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProvider;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProviders;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 
 public class SetAttributesFunction extends LootItemConditionalFunction {
     public static final MapCodec<SetAttributesFunction> MAP_CODEC = RecordCodecBuilder.mapCodec(
@@ -83,7 +_,7 @@
     }
 
     public static SetAttributesFunction.ModifierBuilder modifier(
-        Identifier id, Holder<Attribute> attribute, AttributeModifier.Operation operation, Holder<ContextFloatProvider> amount
+        Identifier id, Holder<Attribute> attribute, AttributeModifier.Operation operation, Holder<NumberProvider> amount
     ) {
         return new SetAttributesFunction.ModifierBuilder(id, attribute, operation, amount);
     }
@@ -120,7 +_,7 @@
     }
 
     private record Modifier(
-        Identifier id, Holder<Attribute> attribute, AttributeModifier.Operation operation, Holder<ContextFloatProvider> amount, List<EquipmentSlotGroup> slots
+        Identifier id, Holder<Attribute> attribute, AttributeModifier.Operation operation, Holder<NumberProvider> amount, List<EquipmentSlotGroup> slots
     ) implements LootContextUser {
         private static final Codec<List<EquipmentSlotGroup>> SLOTS_CODEC = ExtraCodecs.nonEmptyList(ExtraCodecs.compactListCodec(EquipmentSlotGroup.CODEC));
         public static final Codec<SetAttributesFunction.Modifier> CODEC = RecordCodecBuilder.create(
@@ -128,7 +_,7 @@
                     Identifier.CODEC.fieldOf("id").forGetter(SetAttributesFunction.Modifier::id),
                     Attribute.CODEC.fieldOf("attribute").forGetter(SetAttributesFunction.Modifier::attribute),
                     AttributeModifier.Operation.CODEC.fieldOf("operation").forGetter(SetAttributesFunction.Modifier::operation),
-                    ContextFloatProviders.CODEC.fieldOf("amount").forGetter(SetAttributesFunction.Modifier::amount),
+                    NumberProviders.CODEC.fieldOf("amount").forGetter(SetAttributesFunction.Modifier::amount),
                     SLOTS_CODEC.fieldOf("slot").forGetter(SetAttributesFunction.Modifier::slots)
                 )
                 .apply(i, SetAttributesFunction.Modifier::new)
@@ -145,10 +_,10 @@
         private final Identifier id;
         private final Holder<Attribute> attribute;
         private final AttributeModifier.Operation operation;
-        private final Holder<ContextFloatProvider> amount;
+        private final Holder<NumberProvider> amount;
         private final Set<EquipmentSlotGroup> slots = EnumSet.noneOf(EquipmentSlotGroup.class);
 
-        public ModifierBuilder(Identifier id, Holder<Attribute> attribute, AttributeModifier.Operation operation, Holder<ContextFloatProvider> amount) {
+        public ModifierBuilder(Identifier id, Holder<Attribute> attribute, AttributeModifier.Operation operation, Holder<NumberProvider> amount) {
             this.id = id;
             this.attribute = attribute;
             this.operation = operation;
