--- a/net/minecraft/world/level/storage/loot/functions/SetContainerLootTable.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetContainerLootTable.java
@@ -7,8 +_,11 @@
 import java.util.Optional;
 import net.minecraft.core.Holder;
 import net.minecraft.core.component.DataComponents;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.component.SeededContainerLoot;
+import net.minecraft.world.level.block.entity.BlockEntityType;
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.LootTable;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
@@ -21,16 +_,24 @@
         );
     public static final MapCodec<SetContainerLootTable> MAP_CODEC = RecordCodecBuilder.mapCodec(
         i -> commonFields(i)
-            .and(i.group(ID_ONLY_CODEC.fieldOf("loot_table_id").forGetter(f -> f.lootTableId), Codec.LONG.optionalFieldOf("seed", 0L).forGetter(f -> f.seed)))
+            .and(
+                i.group(
+                    ID_ONLY_CODEC.fieldOf("tag").forGetter(f -> f.tag),
+                    Codec.LONG.optionalFieldOf("seed", 0L).forGetter(f -> f.seed),
+                    RegistryFixedCodec.create(Registries.BLOCK_ENTITY_TYPE).fieldOf("type").forGetter(f -> f.type)
+                )
+            )
             .apply(i, SetContainerLootTable::new)
     );
-    private final Holder.Reference<LootTable> lootTableId;
+    private final Holder.Reference<LootTable> tag;
     private final long seed;
+    private final Holder<BlockEntityType<?>> type;
 
-    private SetContainerLootTable(Optional<Holder<LootItemCondition>> condition, Holder.Reference<LootTable> lootTableId, long seed) {
+    private SetContainerLootTable(Optional<Holder<LootItemCondition>> condition, Holder.Reference<LootTable> tag, long seed, Holder<BlockEntityType<?>> type) {
         super(condition);
-        this.lootTableId = lootTableId;
+        this.tag = tag;
         this.seed = seed;
+        this.type = type;
     }
 
     @Override
@@ -44,15 +_,15 @@
             return itemStack;
         }
 
-        itemStack.set(DataComponents.CONTAINER_LOOT, new SeededContainerLoot(this.lootTableId.key(), this.seed));
+        itemStack.set(DataComponents.CONTAINER_LOOT, new SeededContainerLoot(this.tag.key(), this.seed));
         return itemStack;
     }
 
-    public static LootItemConditionalFunction.Builder<?> withLootTable(Holder.Reference<LootTable> value) {
-        return simpleBuilder(conditions -> new SetContainerLootTable(conditions, value, 0L));
+    public static LootItemConditionalFunction.Builder<?> withLootTable(Holder<BlockEntityType<?>> type, Holder.Reference<LootTable> value) {
+        return simpleBuilder(conditions -> new SetContainerLootTable(conditions, value, 0L, type));
     }
 
-    public static LootItemConditionalFunction.Builder<?> withLootTable(Holder.Reference<LootTable> value, long seed) {
-        return simpleBuilder(conditions -> new SetContainerLootTable(conditions, value, seed));
+    public static LootItemConditionalFunction.Builder<?> withLootTable(Holder<BlockEntityType<?>> type, Holder.Reference<LootTable> value, long seed) {
+        return simpleBuilder(conditions -> new SetContainerLootTable(conditions, value, seed, type));
     }
 }
