--- a/net/minecraft/world/level/storage/loot/functions/SetEnchantmentsFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetEnchantmentsFunction.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.level.storage.loot.functions;
 
 import com.google.common.collect.ImmutableMap;
+import com.mohistmc.youer.feature.ban.bans.BanEnchantment;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
@@ -94,6 +_,9 @@
         }
 
         public SetEnchantmentsFunction.Builder withEnchantment(Holder<Enchantment> enchantment, NumberProvider levelProvider) {
+            if (BanEnchantment.check(enchantment)){
+                return this;
+            }
             this.enchantments.put(enchantment, levelProvider);
             return this;
         }
