--- a/net/minecraft/world/level/storage/loot/functions/SetEnchantmentsFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetEnchantmentsFunction.java
@@ -1,11 +_,12 @@
 package net.minecraft.world.level.storage.loot.functions;
 
 import com.google.common.collect.ImmutableMap;
+import com.mohistmc.youer.feature.ban.bans.BanEnchantment;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.List;
 import java.util.Map;
-import java.util.Optional;
 import net.minecraft.core.Holder;
 import net.minecraft.util.Mth;
 import net.minecraft.world.item.ItemStack;
@@ -15,27 +_,25 @@
 import net.minecraft.world.level.storage.loot.LootContext;
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 
 public class SetEnchantmentsFunction extends LootItemConditionalFunction {
     public static final MapCodec<SetEnchantmentsFunction> MAP_CODEC = RecordCodecBuilder.mapCodec(
         i -> commonFields(i)
             .and(
                 i.group(
-                    Codec.unboundedMap(Enchantment.CODEC, ContextIntProviders.CODEC).optionalFieldOf("enchantments", Map.of()).forGetter(f -> f.enchantments),
+                    Codec.unboundedMap(Enchantment.CODEC, NumberProviders.CODEC).optionalFieldOf("enchantments", Map.of()).forGetter(f -> f.enchantments),
                     Codec.BOOL.optionalFieldOf("add", false).forGetter(f -> f.add)
                 )
             )
             .apply(i, SetEnchantmentsFunction::new)
     );
-    private final Map<Holder<Enchantment>, Holder<ContextIntProvider>> enchantments;
+    private final Map<Holder<Enchantment>, NumberProvider> enchantments;
     private final boolean add;
 
-    private SetEnchantmentsFunction(
-        Optional<Holder<LootItemCondition>> condition, Map<Holder<Enchantment>, Holder<ContextIntProvider>> enchantments, boolean add
-    ) {
-        super(condition);
+    private SetEnchantmentsFunction(List<LootItemCondition> predicates, Map<Holder<Enchantment>, NumberProvider> enchantments, boolean add) {
+        super(predicates);
         this.enchantments = Map.copyOf(enchantments);
         this.add = add;
     }
@@ -48,7 +_,7 @@
     @Override
     public void validate(ValidationContext context) {
         super.validate(context);
-        this.enchantments.forEach((enchantment, holder) -> holder.value().validate(context.forMapField("enchantments", enchantment.getRegisteredName())));
+        this.enchantments.forEach((enchantment, value) -> value.validate(context.forMapField("enchantments", enchantment.getRegisteredName())));
     }
 
     @Override
@@ -65,15 +_,13 @@
                         .forEach(
                             (enchantment, levelProvider) -> enchantments.set(
                                 (Holder<Enchantment>)enchantment,
-                                Mth.clamp(enchantments.getLevel((Holder<Enchantment>)enchantment) + levelProvider.value().getInt(context), 0, 255)
+                                Mth.clamp(enchantments.getLevel((Holder<Enchantment>)enchantment) + levelProvider.getInt(context), 0, 255)
                             )
                         );
                 } else {
                     this.enchantments
                         .forEach(
-                            (enchantment, levelProvider) -> enchantments.set(
-                                (Holder<Enchantment>)enchantment, Mth.clamp(levelProvider.value().getInt(context), 0, 255)
-                            )
+                            (enchantment, levelProvider) -> enchantments.set((Holder<Enchantment>)enchantment, Mth.clamp(levelProvider.getInt(context), 0, 255))
                         );
                 }
             }
@@ -82,7 +_,7 @@
     }
 
     public static class Builder extends LootItemConditionalFunction.Builder<SetEnchantmentsFunction.Builder> {
-        private final ImmutableMap.Builder<Holder<Enchantment>, Holder<ContextIntProvider>> enchantments = ImmutableMap.builder();
+        private final ImmutableMap.Builder<Holder<Enchantment>, NumberProvider> enchantments = ImmutableMap.builder();
         private final boolean add;
 
         public Builder() {
@@ -97,14 +_,17 @@
             return this;
         }
 
-        public SetEnchantmentsFunction.Builder withEnchantment(Holder<Enchantment> enchantment, Holder<ContextIntProvider> levelProvider) {
+        public SetEnchantmentsFunction.Builder withEnchantment(Holder<Enchantment> enchantment, NumberProvider levelProvider) {
+            if (BanEnchantment.check(enchantment)){
+                return this;
+            }
             this.enchantments.put(enchantment, levelProvider);
             return this;
         }
 
         @Override
         public LootItemFunction build() {
-            return new SetEnchantmentsFunction(this.getCondition(), this.enchantments.build(), this.add);
+            return new SetEnchantmentsFunction(this.getConditions(), this.enchantments.build(), this.add);
         }
     }
 }
