--- a/net/minecraft/world/level/storage/loot/functions/SetItemDamageFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetItemDamageFunction.java
@@ -12,23 +_,21 @@
 import net.minecraft.world.level.storage.loot.Validatable;
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProvider;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProviders;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 import org.slf4j.Logger;
 
 public class SetItemDamageFunction extends LootItemConditionalFunction {
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final MapCodec<SetItemDamageFunction> MAP_CODEC = RecordCodecBuilder.mapCodec(
         i -> commonFields(i)
-            .and(
-                i.group(ContextFloatProviders.CODEC.fieldOf("damage").forGetter(f -> f.damage), Codec.BOOL.optionalFieldOf("add", false).forGetter(f -> f.add))
-            )
+            .and(i.group(NumberProviders.CODEC.fieldOf("damage").forGetter(f -> f.damage), Codec.BOOL.optionalFieldOf("add", false).forGetter(f -> f.add)))
             .apply(i, SetItemDamageFunction::new)
     );
-    private final Holder<ContextFloatProvider> damage;
+    private final Holder<NumberProvider> damage;
     private final boolean add;
 
-    private SetItemDamageFunction(Optional<Holder<LootItemCondition>> condition, Holder<ContextFloatProvider> damage, boolean add) {
+    private SetItemDamageFunction(Optional<Holder<LootItemCondition>> condition, Holder<NumberProvider> damage, boolean add) {
         super(condition);
         this.damage = damage;
         this.add = add;
@@ -59,11 +_,11 @@
         return itemStack;
     }
 
-    public static LootItemConditionalFunction.Builder<?> setDamage(Holder<ContextFloatProvider> value) {
+    public static LootItemConditionalFunction.Builder<?> setDamage(Holder<NumberProvider> value) {
         return simpleBuilder(conditions -> new SetItemDamageFunction(conditions, value, false));
     }
 
-    public static LootItemConditionalFunction.Builder<?> setDamage(Holder<ContextFloatProvider> value, boolean add) {
+    public static LootItemConditionalFunction.Builder<?> setDamage(Holder<NumberProvider> value, boolean add) {
         return simpleBuilder(conditions -> new SetItemDamageFunction(conditions, value, add));
     }
 }
