--- a/net/minecraft/world/level/storage/loot/functions/SetNameFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetNameFunction.java
@@ -5,7 +_,6 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
