--- a/net/minecraft/world/level/storage/loot/functions/SetOminousBottleAmplifierFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetOminousBottleAmplifierFunction.java
@@ -12,16 +_,16 @@
 import net.minecraft.world.level.storage.loot.Validatable;
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 
 public class SetOminousBottleAmplifierFunction extends LootItemConditionalFunction {
     public static final MapCodec<SetOminousBottleAmplifierFunction> MAP_CODEC = RecordCodecBuilder.mapCodec(
-        i -> commonFields(i).and(ContextIntProviders.CODEC.fieldOf("amplifier").forGetter(f -> f.amplifier)).apply(i, SetOminousBottleAmplifierFunction::new)
+        i -> commonFields(i).and(NumberProviders.CODEC.fieldOf("amplifier").forGetter(f -> f.amplifier)).apply(i, SetOminousBottleAmplifierFunction::new)
     );
-    private final Holder<ContextIntProvider> amplifier;
+    private final Holder<NumberProvider> amplifier;
 
-    private SetOminousBottleAmplifierFunction(Optional<Holder<LootItemCondition>> condition, Holder<ContextIntProvider> amplifier) {
+    private SetOminousBottleAmplifierFunction(Optional<Holder<LootItemCondition>> condition, Holder<NumberProvider> amplifier) {
         super(condition);
         this.amplifier = amplifier;
     }
@@ -44,7 +_,7 @@
         return itemStack;
     }
 
-    public static LootItemConditionalFunction.Builder<?> setAmplifier(Holder<ContextIntProvider> amplifier) {
+    public static LootItemConditionalFunction.Builder<?> setAmplifier(Holder<NumberProvider> amplifier) {
         return simpleBuilder(conditions -> new SetOminousBottleAmplifierFunction(conditions, amplifier));
     }
 }
