--- a/net/minecraft/world/level/storage/loot/functions/SetRandomPotionFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetRandomPotionFunction.java
@@ -5,9 +_,9 @@
 import java.util.Optional;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.alchemy.Potion;
 import net.minecraft.world.item.alchemy.PotionContents;
@@ -17,7 +_,7 @@
 public class SetRandomPotionFunction extends LootItemConditionalFunction {
     public static final MapCodec<SetRandomPotionFunction> MAP_CODEC = RecordCodecBuilder.mapCodec(
         i -> commonFields(i)
-            .and(RegistryCodecs.holderSet(Registries.POTION).optionalFieldOf("options").forGetter(f -> f.options))
+            .and(RegistryCodecs.homogeneousList(Registries.POTION).optionalFieldOf("options").forGetter(f -> f.options))
             .apply(i, SetRandomPotionFunction::new)
     );
     private final Optional<HolderSet<Potion>> options;
