--- a/net/minecraft/world/level/storage/loot/functions/SetStewEffectFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetStewEffectFunction.java
@@ -21,8 +_,8 @@
 import net.minecraft.world.level.storage.loot.Validatable;
 import net.minecraft.world.level.storage.loot.ValidationContext;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 
 public class SetStewEffectFunction extends LootItemConditionalFunction {
     private static final Codec<List<SetStewEffectFunction.EffectEntry>> EFFECTS_LIST = SetStewEffectFunction.EffectEntry.CODEC.listOf().validate(entries -> {
@@ -86,26 +_,22 @@
             return this;
         }
 
-        public SetStewEffectFunction.Builder withEffect(Holder<MobEffect> effect, Holder<ContextIntProvider> duration) {
+        public SetStewEffectFunction.Builder withEffect(Holder<MobEffect> effect, Holder<NumberProvider> duration) {
             this.effects.add(new SetStewEffectFunction.EffectEntry(effect, duration));
             return this;
         }
 
-        public SetStewEffectFunction.Builder withEffect(Holder<MobEffect> effect, int duration) {
-            return this.withEffect(effect, ContextIntProviders.exactly(duration));
-        }
-
         @Override
         public LootItemFunction build() {
             return new SetStewEffectFunction(this.getCondition(), this.effects.build());
         }
     }
 
-    private record EffectEntry(Holder<MobEffect> effect, Holder<ContextIntProvider> duration) implements LootContextUser {
+    private record EffectEntry(Holder<MobEffect> effect, Holder<NumberProvider> duration) implements LootContextUser {
         public static final Codec<SetStewEffectFunction.EffectEntry> CODEC = RecordCodecBuilder.create(
             i -> i.group(
                     MobEffect.CODEC.fieldOf("type").forGetter(SetStewEffectFunction.EffectEntry::effect),
-                    ContextIntProviders.CODEC.fieldOf("duration").forGetter(SetStewEffectFunction.EffectEntry::duration)
+                    NumberProviders.CODEC.fieldOf("duration").forGetter(SetStewEffectFunction.EffectEntry::duration)
                 )
                 .apply(i, SetStewEffectFunction.EffectEntry::new)
         );
