--- a/net/minecraft/world/CompoundContainer.java
+++ b/net/minecraft/world/CompoundContainer.java
@@ -4,8 +_,8 @@
 import net.minecraft.world.item.ItemStack;
 
 public class CompoundContainer implements Container {
-    public final Container container1;
-    public final Container container2;
+    private final Container container1;
+    private final Container container2;
 
     public CompoundContainer(Container p_18913_, Container p_18914_) {
         this.container1 = p_18913_;
