--- a/com/mojang/realmsclient/client/FileDownload.java
+++ b/com/mojang/realmsclient/client/FileDownload.java
@@ -35,6 +_,7 @@
 import net.minecraft.nbt.ReportedNbtException;
 import net.minecraft.util.FileUtil;
 import net.minecraft.util.Util;
+import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.storage.LevelResource;
 import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraft.world.level.validation.ContentValidationException;
@@ -345,7 +_,7 @@
                 file.delete();
             }
 
-            try (LevelStorageSource.LevelStorageAccess access = levelStorageSource.validateAndCreateAccess(finalName)) {
+            try (LevelStorageSource.LevelStorageAccess access = levelStorageSource.validateAndCreateAccess(finalName, LevelStem.OVERWORLD)) {
                 access.renameAndDropPlayer(finalName);
             } catch (NbtException | ReportedNbtException | IOException var40) {
                 LOGGER.error("Failed to modify unpacked realms level {}", finalName, var40);
