--- a/com/mojang/realmsclient/client/RealmsClient.java
+++ b/com/mojang/realmsclient/client/RealmsClient.java
@@ -6,8 +_,6 @@
 import com.mojang.realmsclient.RealmsMainScreen;
 import com.mojang.realmsclient.dto.BackupList;
 import com.mojang.realmsclient.dto.GuardedSerializer;
-import com.mojang.realmsclient.dto.InviteCodeList;
-import com.mojang.realmsclient.dto.InviteCodeRequestDto;
 import com.mojang.realmsclient.dto.Ops;
 import com.mojang.realmsclient.dto.OutboundPlayer;
 import com.mojang.realmsclient.dto.PendingInvite;
@@ -15,7 +_,6 @@
 import com.mojang.realmsclient.dto.PingResult;
 import com.mojang.realmsclient.dto.PlayerInfo;
 import com.mojang.realmsclient.dto.PreferredRegionsDto;
-import com.mojang.realmsclient.dto.RealmTierConfigurationDto;
 import com.mojang.realmsclient.dto.RealmsConfigurationDto;
 import com.mojang.realmsclient.dto.RealmsDescriptionDto;
 import com.mojang.realmsclient.dto.RealmsJoinInformation;
@@ -43,7 +_,6 @@
 import com.mojang.util.UndashedUuid;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.time.Instant;
 import java.util.List;
 import java.util.Locale;
 import java.util.Optional;
@@ -84,14 +_,6 @@
     private static final String TRIALS_RESOURCE = "trial";
     private static final String NOTIFICATIONS_RESOURCE = "notifications";
     private static final String FEATURE_FLAGS_RESOURCE = "feature/v1";
-    private static final String INVITE_CODES_RESOURCE = "api/v1.0/invite/codes";
-    private static final String ACCEPT_HEADER = "Accept";
-    private static final String APPLICATION_JSON = "application/json";
-    private static final String MOJANG_AUTH_HEADER = "x-mojangauth";
-    private static final String REALMS_PLATFORM_HEADER = "X-Realms-Platform";
-    private static final String PATH_INVITE_CODES_UPDATE = "/update";
-    private static final String PATH_INVITE_CODES_DELETE = "/delete/$CODE";
-    private static final String PATH_INVITE_CODES_USE = "/use";
     private static final String PATH_LIST_ALL_REALMS = "/listUserWorldsOfType/any";
     private static final String PATH_CREATE_SNAPSHOT_REALM = "/$PARENT_WORLD_ID/createPrereleaseRealm";
     private static final String PATH_SNAPSHOT_ELIGIBLE_REALMS = "/listPrereleaseEligibleWorlds";
@@ -111,9 +_,6 @@
     private static final String PATH_REJECT_INVITE = "/reject/$INVITATION_ID";
     private static final String PATH_UNINVITE_MYSELF = "/$WORLD_ID";
     private static final String PATH_WORLD_CONFIGURE = "/$WORLD_ID/configuration";
-    private static final String PATH_WORLD_TIER_CONFIGURATION = "/$WORLD_ID/configuration/tier";
-    private static final String PATH_WORLD_RENDER_DISTANCE = "/$WORLD_ID/configuration/renderDistance/$RENDER_DISTANCE";
-    private static final String PATH_WORLD_SIM_DISTANCE = "/$WORLD_ID/configuration/simDistance/$SIM_DISTANCE";
     private static final String PATH_SLOT = "/$WORLD_ID/slot/$SLOT_ID";
     private static final String PATH_WORLD_OPEN = "/$WORLD_ID/open";
     private static final String PATH_WORLD_CLOSE = "/$WORLD_ID/close";
@@ -244,44 +_,6 @@
         return RealmsServer.parse(GSON, json);
     }
 
-    public RealmTierConfigurationDto getRealmTierConfiguration(long realmId) throws RealmsServiceException {
-        String asciiUrl = this.url("worlds" + "/$WORLD_ID/configuration/tier".replace("$WORLD_ID", String.valueOf(realmId)));
-        String json = this.execute(Request.get(asciiUrl));
-
-        try {
-            RealmTierConfigurationDto tierConfiguration = GSON.fromJson(json, RealmTierConfigurationDto.class);
-            if (tierConfiguration != null && tierConfiguration.renderDistance() != null && tierConfiguration.simDistance() != null) {
-                return tierConfiguration;
-            }
-
-            LOGGER.error("Could not parse Realm tier configuration: {}", json);
-        } catch (Exception e) {
-            LOGGER.error("Could not parse Realm tier configuration", e);
-        }
-
-        throw new RealmsServiceException(RealmsError.CustomError.configurationError());
-    }
-
-    public void updateRenderDistance(long realmId, int renderDistance) throws RealmsServiceException {
-        String asciiUrl = this.url(
-            "worlds"
-                + "/$WORLD_ID/configuration/renderDistance/$RENDER_DISTANCE"
-                    .replace("$WORLD_ID", String.valueOf(realmId))
-                    .replace("$RENDER_DISTANCE", String.valueOf(renderDistance))
-        );
-        this.execute(Request.put(asciiUrl, ""));
-    }
-
-    public void updateSimulationDistance(long realmId, int simulationDistance) throws RealmsServiceException {
-        String asciiUrl = this.url(
-            "worlds"
-                + "/$WORLD_ID/configuration/simDistance/$SIM_DISTANCE"
-                    .replace("$WORLD_ID", String.valueOf(realmId))
-                    .replace("$SIM_DISTANCE", String.valueOf(simulationDistance))
-        );
-        this.execute(Request.put(asciiUrl, ""));
-    }
-
     public PreferredRegionsDto getPreferredRegionSelections() throws RealmsServiceException {
         String asciiUrl = this.url("regions/preferredRegions");
         String json = this.execute(Request.get(asciiUrl));
@@ -363,50 +_,6 @@
         return RealmsServer.parse(GSON, json).players;
     }
 
-    public InviteCodeList inviteCodes(long realmId) throws RealmsServiceException {
-        Request<?> request = this.inviteCodesRequest(realmId);
-        String json = this.execute(request);
-        return InviteCodeList.parse(json);
-    }
-
-    private Request<?> inviteCodesRequest(long realmId) {
-        String asciiUrl = frontendV2Url("api/v1.0/invite/codes", "realmId=" + realmId);
-        return frontendV2Request(Request.get(asciiUrl), this.minecraft.getUser().getAccessToken());
-    }
-
-    private static Request<?> frontendV2Request(Request<?> request, String accessToken) {
-        request.connection.setRequestProperty("x-mojangauth", accessToken);
-        if (ENVIRONMENT == RealmsClient.Environment.LOCAL) {
-            request.connection.setRequestProperty("X-Realms-Platform", "Java");
-        }
-
-        return request;
-    }
-
-    public void createInviteCode(long realmId, boolean enabled, @Nullable Instant expiration) throws RealmsServiceException {
-        InviteCodeRequestDto request = new InviteCodeRequestDto(null, realmId, enabled, expiration == null ? null : expiration.toEpochMilli());
-        String asciiUrl = frontendV2Url("api/v1.0/invite/codes", null);
-        this.execute(frontendV2Request(Request.post(asciiUrl, GSON.toJson(request)), this.minecraft.getUser().getAccessToken()));
-    }
-
-    public void updateInviteCode(long realmId, String code, boolean enabled, @Nullable Instant expiration) throws RealmsServiceException {
-        InviteCodeRequestDto request = new InviteCodeRequestDto(code, realmId, enabled, expiration == null ? null : expiration.toEpochMilli());
-        String asciiUrl = frontendV2Url("api/v1.0/invite/codes/update", null);
-        this.execute(frontendV2Request(Request.post(asciiUrl, GSON.toJson(request)), this.minecraft.getUser().getAccessToken()));
-    }
-
-    public void deleteInviteCode(String code) throws RealmsServiceException {
-        String asciiUrl = frontendV2Url("api/v1.0/invite/codes" + "/delete/$CODE".replace("$CODE", code), null);
-        this.execute(frontendV2Request(Request.delete(asciiUrl), this.minecraft.getUser().getAccessToken()));
-    }
-
-    public void redeemInviteCode(String code) throws RealmsServiceException {
-        String asciiUrl = frontendV2Url("api/v1.0/invite/codes/use", null) + "/" + code.trim();
-        Request<?> request = frontendV2Request(Request.post(asciiUrl, ""), this.minecraft.getUser().getAccessToken());
-        request.connection.setRequestProperty("Accept", "application/json");
-        this.execute(request);
-    }
-
     public BackupList backupsFor(long realmId) throws RealmsServiceException {
         String asciiUrl = this.url("worlds" + "/$WORLD_ID/backups".replace("$WORLD_ID", String.valueOf(realmId)));
         String json = this.execute(Request.get(asciiUrl));
@@ -593,20 +_,12 @@
         }
     }
 
-    private static String frontendV2Url(String path, @Nullable String queryString) {
-        try {
-            return new URI(ENVIRONMENT.protocol, ENVIRONMENT.frontendV2Url, "/" + path, queryString, null).toASCIIString();
-        } catch (URISyntaxException e) {
-            throw new IllegalArgumentException(path, e);
-        }
-    }
-
     private String execute(Request<?> request) throws RealmsServiceException {
         request.cookie("sid", this.sessionId);
         request.cookie("user", this.username);
         WorldVersion version = SharedConstants.getCurrentVersion();
         request.cookie("version", version.name());
-        request.addVersionHeaders(version.id(), version.protocolVersion(), RealmsMainScreen.isSnapshot());
+        request.addVersionHeaders(version.protocolVersion(), RealmsMainScreen.isSnapshot());
 
         try {
             int responseCode = request.responseCode();
@@ -645,24 +_,17 @@
     }
 
     public enum Environment {
-        PRODUCTION("pc.realms.minecraft.net", "java.frontendlegacy.realms.minecraft-services.net", "java.frontend.realms.minecraft-services.net", "https"),
-        STAGE(
-            "pc-stage.realms.minecraft.net",
-            "java.frontendlegacy.stage-c2a40e62.realms.minecraft-services.net",
-            "java.frontend.stage-c2a40e62.realms.minecraft-services.net",
-            "https"
-        ),
-        LOCAL("localhost:8080", "localhost:8080", "localhost:8080", "http");
+        PRODUCTION("pc.realms.minecraft.net", "java.frontendlegacy.realms.minecraft-services.net", "https"),
+        STAGE("pc-stage.realms.minecraft.net", "java.frontendlegacy.stage-c2a40e62.realms.minecraft-services.net", "https"),
+        LOCAL("localhost:8080", "localhost:8080", "http");
 
         public final String baseUrl;
         public final String alternativeUrl;
-        public final String frontendV2Url;
         public final String protocol;
 
-        Environment(String baseUrl, String alternativeUrl, String frontendV2Url, String protocol) {
+        Environment(String baseUrl, String alternativeUrl, String protocol) {
             this.baseUrl = baseUrl;
             this.alternativeUrl = alternativeUrl;
-            this.frontendV2Url = frontendV2Url;
             this.protocol = protocol;
         }
 
