--- a/com/mojang/realmsclient/client/Request.java
+++ b/com/mojang/realmsclient/client/Request.java
@@ -23,7 +_,6 @@
     private static final int DEFAULT_READ_TIMEOUT = 60000;
     private static final int DEFAULT_CONNECT_TIMEOUT = 5000;
     private static final String NETWORK_PROTOCOL_KEY = "X-NetworkProtocolVersion";
-    private static final String CLIENT_REF_KEY = "Client-Ref";
     private static final String IS_SNAPSHOT_KEY = "Is-Prerelease";
     private static final String COOKIE_KEY = "Cookie";
 
@@ -59,8 +_,7 @@
         }
     }
 
-    public void addVersionHeaders(String id, int networkProtocol, boolean isSnapshot) {
-        this.connection.addRequestProperty("Client-Ref", id);
+    public void addVersionHeaders(int networkProtocol, boolean isSnapshot) {
         this.connection.addRequestProperty("X-NetworkProtocolVersion", String.valueOf(networkProtocol));
         this.connection.addRequestProperty("Is-Prerelease", String.valueOf(isSnapshot));
     }
