--- a/net/minecraft/tags/BlockTags.java
+++ b/net/minecraft/tags/BlockTags.java
@@ -53,8 +_,6 @@
     public static final TagKey<Block> TERRACOTTA = BlockItemTags.TERRACOTTA.block();
     public static final TagKey<Block> GLAZED_TERRACOTTA = BlockItemTags.GLAZED_TERRACOTTA.block();
     public static final TagKey<Block> CONCRETE = BlockItemTags.CONCRETE.block();
-    public static final TagKey<Block> CONCRETE_SLABS = BlockItemTags.CONCRETE_SLABS.block();
-    public static final TagKey<Block> CONCRETE_STAIRS = BlockItemTags.CONCRETE_STAIRS.block();
     public static final TagKey<Block> CONCRETE_POWDERS = BlockItemTags.CONCRETE_POWDERS.block();
     public static final TagKey<Block> COMPLETES_FIND_TREE_TUTORIAL = BlockItemTags.COMPLETES_FIND_TREE_TUTORIAL.block();
     public static final TagKey<Block> SHULKER_BOXES = BlockItemTags.SHULKER_BOXES.block();
@@ -118,22 +_,18 @@
     public static final TagKey<Block> CAN_GLIDE_THROUGH = create("can_glide_through");
     public static final TagKey<Block> CAUSES_PERIODIC_GEYSER_ERUPTIONS = create("causes_periodic_geyser_eruptions");
     public static final TagKey<Block> CAUSES_CONTINUOUS_GEYSER_ERUPTIONS = create("causes_continuous_geyser_eruptions");
-    public static final TagKey<Block> NETHER_PORTAL_FRAME = create("nether_portal_frame");
-    public static final TagKey<Block> CONDUIT_EFFECT_BLOCK = create("conduit_effect_block");
     public static final TagKey<Block> INFINIBURN_OVERWORLD = create("infiniburn_overworld");
     public static final TagKey<Block> INFINIBURN_NETHER = create("infiniburn_nether");
     public static final TagKey<Block> INFINIBURN_END = create("infiniburn_end");
     public static final TagKey<Block> SUBSTRATE_OVERWORLD = create("substrate_overworld");
     public static final TagKey<Block> BASE_STONE_OVERWORLD = create("base_stone_overworld");
     public static final TagKey<Block> STONE_ORE_REPLACEABLES = create("stone_ore_replaceables");
-    public static final TagKey<Block> HEIGHT_SPECIFIC_ORE_REPLACEABLES = create("height_specific_ore_replaceables");
     public static final TagKey<Block> DEEPSLATE_ORE_REPLACEABLES = create("deepslate_ore_replaceables");
     public static final TagKey<Block> BASE_STONE_NETHER = create("base_stone_nether");
     public static final TagKey<Block> BENEATH_TREE_PODZOL_REPLACEABLE = create("beneath_tree_podzol_replaceable");
     public static final TagKey<Block> BENEATH_BAMBOO_PODZOL_REPLACEABLE = create("beneath_bamboo_podzol_replaceable");
     public static final TagKey<Block> CANNOT_REPLACE_BELOW_TREE_TRUNK = create("cannot_replace_below_tree_trunk");
     public static final TagKey<Block> CANNOT_PLACE_BASALT_PILLAR_ON = create("cannot_place_basalt_pillar_on");
-    public static final TagKey<Block> UNCARVABLE = create("uncarvable");
     public static final TagKey<Block> CANDLE_CAKES = create("candle_cakes");
     public static final TagKey<Block> CAULDRONS = create("cauldrons");
     public static final TagKey<Block> CRYSTAL_SOUND_BLOCKS = create("crystal_sound_blocks");
@@ -142,7 +_,6 @@
     public static final TagKey<Block> CAMEL_SAND_STEP_SOUND_BLOCKS = create("camel_sand_step_sound_blocks");
     public static final TagKey<Block> HAPPY_GHAST_AVOIDS = create("happy_ghast_avoids");
     public static final TagKey<Block> OCCLUDES_VIBRATION_SIGNALS = create("occludes_vibration_signals");
-    public static final TagKey<Block> CUSHION_USES_COLLISION_SHAPE = create("cushion_uses_collision_shape");
     public static final TagKey<Block> DRIPSTONE_REPLACEABLE = create("dripstone_replaceable_blocks");
     public static final TagKey<Block> SULFUR_SPIKE_REPLACEABLE = create("sulfur_spike_replaceable_blocks");
     public static final TagKey<Block> CAVE_VINES = create("cave_vines");
@@ -179,7 +_,6 @@
     public static final TagKey<Block> LAVA_POOL_STONE_CANNOT_REPLACE = create("lava_pool_stone_cannot_replace");
     public static final TagKey<Block> GEODE_INVALID_BLOCKS = create("geode_invalid_blocks");
     public static final TagKey<Block> FROG_PREFER_JUMP_TO = create("frog_prefer_jump_to");
-    public static final TagKey<Block> SCULK_GROWTH_INHIBITORS = create("sculk_growth_inhibitors");
     public static final TagKey<Block> SCULK_REPLACEABLE = create("sculk_replaceable");
     public static final TagKey<Block> SCULK_REPLACEABLE_WORLD_GEN = create("sculk_replaceable_world_gen");
     public static final TagKey<Block> ANCIENT_CITY_REPLACEABLE = create("ancient_city_replaceable");
@@ -198,7 +_,7 @@
     public static final TagKey<Block> BATS_SPAWNABLE_ON = create("bats_spawnable_on");
     public static final TagKey<Block> CAMELS_SPAWNABLE_ON = create("camels_spawnable_on");
     public static final TagKey<Block> AZALEA_GROWS_ON = create("azalea_grows_on");
-    public static final TagKey<Block> CONVERTIBLE_TO_MUD = create("convertible_to_mud");
+    public static final TagKey<Block> CONVERTABLE_TO_MUD = create("convertable_to_mud");
     public static final TagKey<Block> MANGROVE_LOGS_CAN_GROW_THROUGH = create("mangrove_logs_can_grow_through");
     public static final TagKey<Block> MANGROVE_ROOTS_CAN_GROW_THROUGH = create("mangrove_roots_can_grow_through");
     public static final TagKey<Block> SNAPS_GOAT_HORN = create("snaps_goat_horn");
@@ -206,11 +_,6 @@
     public static final TagKey<Block> REPLACEABLE_BY_MUSHROOMS = create("replaceable_by_mushrooms");
     public static final TagKey<Block> ENABLES_BUBBLE_COLUMN_DRAG_DOWN = create("enables_bubble_column_drag_down");
     public static final TagKey<Block> ENABLES_BUBBLE_COLUMN_PUSH_UP = create("enables_bubble_column_push_up");
-    public static final TagKey<Block> CATS_CAN_SIT_ON = create("cats_can_sit_on");
-    public static final TagKey<Block> CATS_CAN_LIE_ON = create("cats_can_lie_on");
-    public static final TagKey<Block> SPEEDS_UP_ZOMBIE_VILLAGER_CURING = create("speeds_up_zombie_villager_curing");
-    public static final TagKey<Block> VILLAGERS_CAN_SLEEP_ON_BED = create("villagers_can_sleep_on_bed");
-    public static final TagKey<Block> VILLAGER_BABIES_CAN_JUMP_ON_BED = create("villager_babies_can_jump_on_bed");
     public static final TagKey<Block> SUPPORTS_VEGETATION = create("supports_vegetation");
     public static final TagKey<Block> SUPPORTS_DRY_VEGETATION = create("supports_dry_vegetation");
     public static final TagKey<Block> SUPPORTS_CROPS = create("supports_crops");
@@ -265,11 +_,6 @@
     public static final TagKey<Block> TRIGGERS_AMBIENT_DRIED_GHAST_BLOCK_SOUNDS = create("triggers_ambient_dried_ghast_block_sounds");
     public static final TagKey<Block> REQUIRED_FOR_POPLAR_LEAF_AMBIENCE = create("required_for_poplar_leaf_ambience");
     public static final TagKey<Block> SPELEOTHEMS = create("speleothems");
-    public static final TagKey<Block> DANGEROUS_FOR_TELEPORTATION = create("dangerous_for_teleportation");
-    public static final TagKey<Block> CAT_DOES_NOT_TELEPORT_TO = create("cat_does_not_teleport_to");
-    public static final TagKey<Block> ENDERMAN_DOES_NOT_TELEPORT_TO = create("enderman_does_not_teleport_to");
-    public static final TagKey<Block> SHULKER_DOES_NOT_TELEPORT_TO = create("shulker_does_not_teleport_to");
-    public static final TagKey<Block> CONSUMABLE_DOES_NOT_TELEPORT_TO = create("consumable_does_not_teleport_to");
     public static final TagKey<Block> FOX_IMMUNE_TO = create("fox_immune_to");
     public static final TagKey<Block> POLAR_BEAR_IMMUNE_TO = create("polar_bear_immune_to");
     public static final TagKey<Block> SNOW_GOLEM_IMMUNE_TO = create("snow_golem_immune_to");
@@ -295,5 +_,9 @@
 
     private static TagKey<Block> create(String name) {
         return TagKey.create(Registries.BLOCK, Identifier.withDefaultNamespace(name));
+    }
+
+    public static TagKey<Block> create(Identifier name) {
+        return TagKey.create(Registries.BLOCK, name);
     }
 }
