--- a/net/minecraft/tags/BlockTags.java
+++ b/net/minecraft/tags/BlockTags.java
@@ -122,7 +_,6 @@
     public static final TagKey<Block> SUBSTRATE_OVERWORLD = create("substrate_overworld");
     public static final TagKey<Block> BASE_STONE_OVERWORLD = create("base_stone_overworld");
     public static final TagKey<Block> STONE_ORE_REPLACEABLES = create("stone_ore_replaceables");
-    public static final TagKey<Block> HEIGHT_SPECIFIC_ORE_REPLACEABLES = create("height_specific_ore_replaceables");
     public static final TagKey<Block> DEEPSLATE_ORE_REPLACEABLES = create("deepslate_ore_replaceables");
     public static final TagKey<Block> BASE_STONE_NETHER = create("base_stone_nether");
     public static final TagKey<Block> BENEATH_TREE_PODZOL_REPLACEABLE = create("beneath_tree_podzol_replaceable");
@@ -173,7 +_,6 @@
     public static final TagKey<Block> LAVA_POOL_STONE_CANNOT_REPLACE = create("lava_pool_stone_cannot_replace");
     public static final TagKey<Block> GEODE_INVALID_BLOCKS = create("geode_invalid_blocks");
     public static final TagKey<Block> FROG_PREFER_JUMP_TO = create("frog_prefer_jump_to");
-    public static final TagKey<Block> SCULK_GROWTH_INHIBITORS = create("sculk_growth_inhibitors");
     public static final TagKey<Block> SCULK_REPLACEABLE = create("sculk_replaceable");
     public static final TagKey<Block> SCULK_REPLACEABLE_WORLD_GEN = create("sculk_replaceable_world_gen");
     public static final TagKey<Block> ANCIENT_CITY_REPLACEABLE = create("ancient_city_replaceable");
@@ -192,7 +_,7 @@
     public static final TagKey<Block> BATS_SPAWNABLE_ON = create("bats_spawnable_on");
     public static final TagKey<Block> CAMELS_SPAWNABLE_ON = create("camels_spawnable_on");
     public static final TagKey<Block> AZALEA_GROWS_ON = create("azalea_grows_on");
-    public static final TagKey<Block> CONVERTIBLE_TO_MUD = create("convertible_to_mud");
+    public static final TagKey<Block> CONVERTABLE_TO_MUD = create("convertable_to_mud");
     public static final TagKey<Block> MANGROVE_LOGS_CAN_GROW_THROUGH = create("mangrove_logs_can_grow_through");
     public static final TagKey<Block> MANGROVE_ROOTS_CAN_GROW_THROUGH = create("mangrove_roots_can_grow_through");
     public static final TagKey<Block> SNAPS_GOAT_HORN = create("snaps_goat_horn");
@@ -200,11 +_,6 @@
     public static final TagKey<Block> REPLACEABLE_BY_MUSHROOMS = create("replaceable_by_mushrooms");
     public static final TagKey<Block> ENABLES_BUBBLE_COLUMN_DRAG_DOWN = create("enables_bubble_column_drag_down");
     public static final TagKey<Block> ENABLES_BUBBLE_COLUMN_PUSH_UP = create("enables_bubble_column_push_up");
-    public static final TagKey<Block> CATS_CAN_SIT_ON = create("cats_can_sit_on");
-    public static final TagKey<Block> CATS_CAN_LIE_ON = create("cats_can_lie_on");
-    public static final TagKey<Block> SPEEDS_UP_ZOMBIE_VILLAGER_CURING = create("speeds_up_zombie_villager_curing");
-    public static final TagKey<Block> VILLAGERS_CAN_SLEEP_ON_BED = create("villagers_can_sleep_on_bed");
-    public static final TagKey<Block> VILLAGER_BABIES_CAN_JUMP_ON_BED = create("villager_babies_can_jump_on_bed");
     public static final TagKey<Block> SUPPORTS_VEGETATION = create("supports_vegetation");
     public static final TagKey<Block> SUPPORTS_DRY_VEGETATION = create("supports_dry_vegetation");
     public static final TagKey<Block> SUPPORTS_CROPS = create("supports_crops");
@@ -284,5 +_,9 @@
 
     private static TagKey<Block> create(String name) {
         return TagKey.create(Registries.BLOCK, Identifier.withDefaultNamespace(name));
+    }
+
+    public static TagKey<Block> create(Identifier name) {
+        return TagKey.create(Registries.BLOCK, name);
     }
 }
