--- a/net/minecraft/tags/BlockItemTags.java
+++ b/net/minecraft/tags/BlockItemTags.java
@@ -65,8 +_,6 @@
     public static final BlockItemTagId TERRACOTTA = BlockItemTagId.create("terracotta");
     public static final BlockItemTagId GLAZED_TERRACOTTA = BlockItemTagId.create("glazed_terracotta");
     public static final BlockItemTagId CONCRETE = BlockItemTagId.create("concrete");
-    public static final BlockItemTagId CONCRETE_STAIRS = BlockItemTagId.create("concrete_stairs");
-    public static final BlockItemTagId CONCRETE_SLABS = BlockItemTagId.create("concrete_slabs");
     public static final BlockItemTagId CONCRETE_POWDERS = BlockItemTagId.create("concrete_powders");
     public static final BlockItemTagId COMPLETES_FIND_TREE_TUTORIAL = BlockItemTagId.create("completes_find_tree_tutorial");
     public static final BlockItemTagId SHULKER_BOXES = BlockItemTagId.create("shulker_boxes");
