--- a/net/minecraft/tags/ItemTags.java
+++ b/net/minecraft/tags/ItemTags.java
@@ -88,6 +_,7 @@
     public static final TagKey<Item> IGNORED_BY_PIGLIN_BABIES = bind("ignored_by_piglin_babies");
     public static final TagKey<Item> PIGLIN_SAFE_ARMOR = bind("piglin_safe_armor");
     public static final TagKey<Item> DUPLICATES_ALLAYS = bind("duplicates_allays");
+    public static final TagKey<Item> BREWING_FUEL = bind("brewing_fuel");
     public static final TagKey<Item> EGGS = bind("eggs");
     public static final TagKey<Item> MEAT = bind("meat");
     public static final TagKey<Item> SNIFFER_FOOD = bind("sniffer_food");
@@ -186,7 +_,6 @@
     public static final TagKey<Item> CAULDRON_CAN_REMOVE_DYE = bind("cauldron_can_remove_dye");
     public static final TagKey<Item> CAT_COLLAR_DYES = bind("cat_collar_dyes");
     public static final TagKey<Item> WOLF_COLLAR_DYES = bind("wolf_collar_dyes");
-    public static final TagKey<Item> CUSHIONS = bind("cushions");
     public static final TagKey<Item> SKELETON_PREFERRED_WEAPONS = bind("skeleton_preferred_weapons");
     public static final TagKey<Item> DROWNED_PREFERRED_WEAPONS = bind("drowned_preferred_weapons");
     public static final TagKey<Item> PIGLIN_PREFERRED_WEAPONS = bind("piglin_preferred_weapons");
@@ -194,7 +_,7 @@
     public static final TagKey<Item> WITHER_SKELETON_DISLIKED_WEAPONS = bind("wither_skeleton_disliked_weapons");
     public static final TagKey<Item> SHEARABLE_FROM_COPPER_GOLEM = bind("shearable_from_copper_golem");
     public static final TagKey<Item> METAL_NUGGETS = bind("metal_nuggets");
-    public static final TagKey<Item> DOUSES_CAMPFIRES = bind("douses_campfires");
+    public static final TagKey<Item> DOUSES_CAMPFIRES = bind("dowses_campfires");
     public static final TagKey<Item> FOOT_ARMOR_ENCHANTABLE = bind("enchantable/foot_armor");
     public static final TagKey<Item> LEG_ARMOR_ENCHANTABLE = bind("enchantable/leg_armor");
     public static final TagKey<Item> CHEST_ARMOR_ENCHANTABLE = bind("enchantable/chest_armor");
@@ -224,5 +_,9 @@
 
     private static TagKey<Item> bind(String name) {
         return TagKey.create(Registries.ITEM, Identifier.withDefaultNamespace(name));
+    }
+
+    public static TagKey<Item> create(final Identifier name) {
+        return TagKey.create(Registries.ITEM, name);
     }
 }
