--- a/net/minecraft/tags/ItemTags.java
+++ b/net/minecraft/tags/ItemTags.java
@@ -61,8 +_,6 @@
     public static final TagKey<Item> TERRACOTTA = BlockItemTags.TERRACOTTA.item();
     public static final TagKey<Item> GLAZED_TERRACOTTA = BlockItemTags.GLAZED_TERRACOTTA.item();
     public static final TagKey<Item> CONCRETE = BlockItemTags.CONCRETE.item();
-    public static final TagKey<Item> CONCRETE_STAIRS = BlockItemTags.CONCRETE_STAIRS.item();
-    public static final TagKey<Item> CONCRETE_SLABS = BlockItemTags.CONCRETE_SLABS.item();
     public static final TagKey<Item> CONCRETE_POWDERS = BlockItemTags.CONCRETE_POWDERS.item();
     public static final TagKey<Item> COMPLETES_FIND_TREE_TUTORIAL = BlockItemTags.COMPLETES_FIND_TREE_TUTORIAL.item();
     public static final TagKey<Item> SHULKER_BOXES = BlockItemTags.SHULKER_BOXES.item();
@@ -90,6 +_,7 @@
     public static final TagKey<Item> IGNORED_BY_PIGLIN_BABIES = bind("ignored_by_piglin_babies");
     public static final TagKey<Item> PIGLIN_SAFE_ARMOR = bind("piglin_safe_armor");
     public static final TagKey<Item> DUPLICATES_ALLAYS = bind("duplicates_allays");
+    public static final TagKey<Item> BREWING_FUEL = bind("brewing_fuel");
     public static final TagKey<Item> EGGS = bind("eggs");
     public static final TagKey<Item> MEAT = bind("meat");
     public static final TagKey<Item> SNIFFER_FOOD = bind("sniffer_food");
@@ -160,8 +_,6 @@
     public static final TagKey<Item> FREEZE_IMMUNE_WEARABLES = bind("freeze_immune_wearables");
     public static final TagKey<Item> CLUSTER_MAX_HARVESTABLES = bind("cluster_max_harvestables");
     public static final TagKey<Item> COMPASSES = bind("compasses");
-    public static final TagKey<Item> CLONABLE_MAPS = bind("clonable_maps");
-    public static final TagKey<Item> EXTENDABLE_MAPS = bind("extendable_maps");
     public static final TagKey<Item> CREEPER_IGNITERS = bind("creeper_igniters");
     public static final TagKey<Item> NOTE_BLOCK_TOP_INSTRUMENTS = bind("noteblock_top_instruments");
     public static final TagKey<Item> FOOT_ARMOR = bind("foot_armor");
@@ -190,7 +_,6 @@
     public static final TagKey<Item> CAULDRON_CAN_REMOVE_DYE = bind("cauldron_can_remove_dye");
     public static final TagKey<Item> CAT_COLLAR_DYES = bind("cat_collar_dyes");
     public static final TagKey<Item> WOLF_COLLAR_DYES = bind("wolf_collar_dyes");
-    public static final TagKey<Item> CUSHIONS = bind("cushions");
     public static final TagKey<Item> SKELETON_PREFERRED_WEAPONS = bind("skeleton_preferred_weapons");
     public static final TagKey<Item> DROWNED_PREFERRED_WEAPONS = bind("drowned_preferred_weapons");
     public static final TagKey<Item> PIGLIN_PREFERRED_WEAPONS = bind("piglin_preferred_weapons");
@@ -198,9 +_,7 @@
     public static final TagKey<Item> WITHER_SKELETON_DISLIKED_WEAPONS = bind("wither_skeleton_disliked_weapons");
     public static final TagKey<Item> SHEARABLE_FROM_COPPER_GOLEM = bind("shearable_from_copper_golem");
     public static final TagKey<Item> METAL_NUGGETS = bind("metal_nuggets");
-    public static final TagKey<Item> DOUSES_CAMPFIRES = bind("douses_campfires");
-    public static final TagKey<Item> BREWING_POTION_INPUTS = bind("brewing_potion_inputs");
-    public static final TagKey<Item> FURNACE_FUEL_BOTTOM_TAKEABLE = bind("furnace_fuel_bottom_takeable");
+    public static final TagKey<Item> DOUSES_CAMPFIRES = bind("dowses_campfires");
     public static final TagKey<Item> FOOT_ARMOR_ENCHANTABLE = bind("enchantable/foot_armor");
     public static final TagKey<Item> LEG_ARMOR_ENCHANTABLE = bind("enchantable/leg_armor");
     public static final TagKey<Item> CHEST_ARMOR_ENCHANTABLE = bind("enchantable/chest_armor");
@@ -230,5 +_,9 @@
 
     private static TagKey<Item> bind(String name) {
         return TagKey.create(Registries.ITEM, Identifier.withDefaultNamespace(name));
+    }
+
+    public static TagKey<Item> create(final Identifier name) {
+        return TagKey.create(Registries.ITEM, name);
     }
 }
