--- a/net/minecraft/tags/PotionTags.java
+++ b/net/minecraft/tags/PotionTags.java
@@ -6,10 +_,6 @@
 
 public class PotionTags {
     public static final TagKey<Potion> TRADEABLE = create("tradeable");
-    public static final TagKey<Potion> DOUSES_FIRE = create("douses_fire");
-    public static final TagKey<Potion> HURTS_WATER_SENSITIVE_ENTITIES = create("hurts_water_sensitive_entities");
-    public static final TagKey<Potion> EXTINGUISHES_ENTITIES = create("extinguishes_entities");
-    public static final TagKey<Potion> REHYDRATES_AXOLOTLS = create("rehydrates_axolotls");
 
     private PotionTags() {
     }
