--- a/net/minecraft/tags/StructureTags.java
+++ b/net/minecraft/tags/StructureTags.java
@@ -37,7 +_,7 @@
     TagKey<Structure> ON_ABANDONED_CAMP_FLOWER_FOREST_MAPS = create("on_abandoned_camp_flower_forest");
     TagKey<Structure> ON_ABANDONED_CAMP_PALE_GARDEN_MAPS = create("on_abandoned_camp_pale_garden");
     TagKey<Structure> ON_ABANDONED_CAMP_SWAMP_MAPS = create("on_abandoned_camp_swamp");
-    TagKey<Structure> ON_ABANDONED_CAMP_WINDSWEPT_FOREST_MAPS = create("on_abandoned_camp_windswept_forest");
+    TagKey<Structure> ON_ABANDONED_CAMP_WINDSWEPT_FOREST_MAPS = create("on_abandoned_camp_windswept");
 
     private static TagKey<Structure> create(String name) {
         return TagKey.create(Registries.STRUCTURE, Identifier.withDefaultNamespace(name));
