--- a/net/minecraft/world/level/levelgen/placement/NoiseThresholdCountPlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/NoiseThresholdCountPlacement.java
@@ -23,7 +_,7 @@
 
     @Override
     public int count(RandomSource random, BlockPos origin) {
-        double flowerNoise = Biome.BIOME_INFO_NOISE.get(origin.getX() / 200.0, origin.getZ() / 200.0);
+        double flowerNoise = Biome.BIOME_INFO_NOISE.getValue(origin.getX() / 200.0, origin.getZ() / 200.0, false);
         return flowerNoise < this.noiseLevel ? this.belowNoise : this.aboveNoise;
     }
 
