--- a/net/minecraft/world/level/levelgen/placement/NoiseBasedCountPlacement.java
+++ b/net/minecraft/world/level/levelgen/placement/NoiseBasedCountPlacement.java
@@ -23,7 +_,7 @@
 
     @Override
     public int count(RandomSource random, BlockPos origin) {
-        double flowerNoise = Biome.BIOME_INFO_NOISE.get(origin.getX() / this.noiseFactor, origin.getZ() / this.noiseFactor);
+        double flowerNoise = Biome.BIOME_INFO_NOISE.getValue(origin.getX() / this.noiseFactor, origin.getZ() / this.noiseFactor, false);
         return (int)Math.ceil((flowerNoise + this.noiseOffset) * this.noiseToCountRatio);
     }
 
