--- a/net/minecraft/world/level/levelgen/placement/PlacedFeature.java
+++ b/net/minecraft/world/level/levelgen/placement/PlacedFeature.java
@@ -7,8 +_,9 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
+import net.minecraft.resources.RegistryFileCodec;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.chunk.ChunkGenerator;
@@ -21,9 +_,11 @@
             )
             .apply(i, PlacedFeature::new)
     );
-    public static final Codec<Holder<PlacedFeature>> CODEC = RegistryCodecs.holder(Registries.PLACED_FEATURE, DIRECT_CODEC);
-    public static final Codec<HolderSet<PlacedFeature>> LIST_CODEC = RegistryCodecs.holderSet(Registries.PLACED_FEATURE, DIRECT_CODEC);
-    public static final Codec<List<HolderSet<PlacedFeature>>> LIST_OF_LISTS_CODEC = RegistryCodecs.holderSet(Registries.PLACED_FEATURE, DIRECT_CODEC, true)
+    public static final Codec<Holder<PlacedFeature>> CODEC = RegistryFileCodec.create(Registries.PLACED_FEATURE, DIRECT_CODEC);
+    public static final Codec<HolderSet<PlacedFeature>> LIST_CODEC = RegistryCodecs.homogeneousList(Registries.PLACED_FEATURE, DIRECT_CODEC);
+    public static final Codec<List<HolderSet<PlacedFeature>>> LIST_OF_LISTS_CODEC = RegistryCodecs.homogeneousList(
+            Registries.PLACED_FEATURE, DIRECT_CODEC, true
+        )
         .listOf();
 
     public boolean place(WorldGenLevel level, ChunkGenerator generator, RandomSource random, BlockPos origin) {
