--- a/net/minecraft/world/level/levelgen/placement/PlacementModifierTypes.java
+++ b/net/minecraft/world/level/levelgen/placement/PlacementModifierTypes.java
@@ -22,7 +_,6 @@
         Registry.register(registry, "height_range", HeightRangePlacement.CODEC);
         Registry.register(registry, "in_square", InSquarePlacement.CODEC);
         Registry.register(registry, "offset", OffsetPlacement.CODEC);
-        Registry.register(registry, "randomly_selected", RandomlySelectedPlacement.CODEC);
         return Registry.register(registry, "fixed_placement", FixedPlacement.CODEC);
     }
 }
