--- a/net/minecraft/world/entity/animal/chicken/ChickenSoundVariant.java
+++ b/net/minecraft/world/entity/animal/chicken/ChickenSoundVariant.java
@@ -4,16 +_,16 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.sounds.SoundEvent;
 
 public record ChickenSoundVariant(ChickenSoundVariant.ChickenSoundSet adultSounds, ChickenSoundVariant.ChickenSoundSet babySounds) {
     public static final Codec<ChickenSoundVariant> DIRECT_CODEC = codec();
     public static final Codec<ChickenSoundVariant> NETWORK_CODEC = codec();
-    public static final Codec<Holder<ChickenSoundVariant>> CODEC = RegistryCodecs.holder(Registries.CHICKEN_SOUND_VARIANT);
+    public static final Codec<Holder<ChickenSoundVariant>> CODEC = RegistryFixedCodec.create(Registries.CHICKEN_SOUND_VARIANT);
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<ChickenSoundVariant>> STREAM_CODEC = ByteBufCodecs.holderRegistry(
         Registries.CHICKEN_SOUND_VARIANT
     );
