--- a/net/minecraft/world/entity/animal/nautilus/AbstractNautilus.java
+++ b/net/minecraft/world/entity/animal/nautilus/AbstractNautilus.java
@@ -481,7 +_,7 @@
 
     protected void createInventory() {
         SimpleContainer old = this.inventory;
-        this.inventory = new SimpleContainer(this.getInventorySize());
+        this.inventory = new SimpleContainer(this.getInventorySize(), (org.bukkit.entity.AbstractNautilus) this.getBukkitEntity()); // CraftBukkit - store entity
         if (old != null) {
             int max = Math.min(old.getContainerSize(), this.inventory.getContainerSize());
 
