--- a/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
+++ b/net/minecraft/world/entity/animal/nautilus/ZombieNautilus.java
@@ -46,6 +_,7 @@
 
     public ZombieNautilus(EntityType<? extends ZombieNautilus> type, Level level) {
         super(type, level);
+        this.setShouldBurnInDay(true); // Purpur - API for any mob to burn daylight
     }
 
     public static AttributeSupplier.Builder createAttributes() {
