--- a/net/minecraft/world/entity/animal/nautilus/ZombieNautilusVariant.java
+++ b/net/minecraft/world/entity/animal/nautilus/ZombieNautilusVariant.java
@@ -5,10 +_,10 @@
 import java.util.List;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.util.StringRepresentable;
 import net.minecraft.world.entity.variant.ModelAndTexture;
 import net.minecraft.world.entity.variant.PriorityProvider;
@@ -33,7 +_,7 @@
             )
             .apply(i, ZombieNautilusVariant::new)
     );
-    public static final Codec<Holder<ZombieNautilusVariant>> CODEC = RegistryCodecs.holder(Registries.ZOMBIE_NAUTILUS_VARIANT);
+    public static final Codec<Holder<ZombieNautilusVariant>> CODEC = RegistryFixedCodec.create(Registries.ZOMBIE_NAUTILUS_VARIANT);
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<ZombieNautilusVariant>> STREAM_CODEC = ByteBufCodecs.holderRegistry(
         Registries.ZOMBIE_NAUTILUS_VARIANT
     );
