--- a/com/mojang/renderpearl/api/device/DeviceLimits.java
+++ b/com/mojang/renderpearl/api/device/DeviceLimits.java
@@ -11,8 +_,7 @@
     int maxTextureSize,
     long maxMemoryAllocationSize,
     int maxMultiDrawDirectInterleavedDrawCount,
-    int maxColorAttachments,
-    int maxDrawIndirectDrawCount
+    int maxColorAttachments
 ) {
     public int maxTextureSizeForFormat(GpuFormat format) {
         return Integer.highestOneBit(Math.min(this.maxTextureSize, (int)Math.sqrt((double)this.maxMemoryAllocationSize / format.blockSize())));
