--- a/com/mojang/renderpearl/api/device/GpuBackend.java
+++ b/com/mojang/renderpearl/api/device/GpuBackend.java
@@ -12,7 +_,11 @@
 
     void unloadLibrary();
 
-    long createWindow(@Nullable String title, int width, int height, long flags);
-
-    GpuDevice createDevice(GpuDebugOptions debugOptions) throws BackendCreationException;
+    void setWindowHints();
+
+    long getWindowFlags();
+
+    void handleWindowCreationErrors(final @Nullable String error) throws BackendCreationException;
+
+    GpuDevice createDevice(long window, GpuDebugOptions debugOptions) throws BackendCreationException;
 }
