--- a/com/mojang/renderpearl/api/device/GpuDevice.java
+++ b/com/mojang/renderpearl/api/device/GpuDevice.java
@@ -15,9 +_,6 @@
 import java.nio.ByteBuffer;
 import java.util.List;
 import java.util.OptionalDouble;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.Executor;
-import java.util.function.BooleanSupplier;
 import java.util.function.Supplier;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
@@ -25,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public interface GpuDevice {
-    GpuSurface createSurface(long windowHandle, final BooleanSupplier isIconified);
+    GpuSurface createSurface(long windowHandle);
 
     CommandEncoder createCommandEncoder();
 
@@ -51,7 +_,7 @@
 
     boolean isDebuggingEnabled();
 
-    CompletableFuture<CompiledRenderPipeline.Pending> compilePipeline(RenderPipeline pipeline, ShaderSource shaderSource, Executor executor);
+    @Nullable CompiledRenderPipeline compilePipeline(RenderPipeline pipeline, ShaderSource shaderSource);
 
     void close();
 
