--- a/com/mojang/renderpearl/api/device/GpuSurface.java
+++ b/com/mojang/renderpearl/api/device/GpuSurface.java
@@ -2,14 +_,16 @@
 
 import com.mojang.renderpearl.api.commands.CommandEncoder;
 import com.mojang.renderpearl.api.textures.GpuTextureView;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.util.Collection;
 import java.util.Optional;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface GpuSurface extends UncheckedAutoCloseable {
+public interface GpuSurface extends AutoCloseable {
+    @Override
+    void close();
+
     void configure(GpuSurface.Configuration config) throws SurfaceException;
 
     Optional<GpuSurface.Configuration> currentConfiguration();
@@ -35,7 +_,7 @@
         FIFO,
         FIFO_RELAXED;
 
-        private static final GpuSurface.PresentMode[] PRESENT_MODES_VSYNC = new GpuSurface.PresentMode[]{FIFO};
+        private static final GpuSurface.PresentMode[] PRESENT_MODES_VSYNC = new GpuSurface.PresentMode[]{FIFO_RELAXED, FIFO};
         private static final GpuSurface.PresentMode[] PRESENT_MODES_NO_VSYNC = new GpuSurface.PresentMode[]{IMMEDIATE, MAILBOX, FIFO};
 
         public static GpuSurface.PresentMode getSupportedVsyncMode(Collection<GpuSurface.PresentMode> supportedModes, boolean vsync) {
