--- a/com/mojang/renderpearl/api/device/HintsAndWorkarounds.java
+++ b/com/mojang/renderpearl/api/device/HintsAndWorkarounds.java
@@ -4,7 +_,5 @@
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public record HintsAndWorkarounds(
-    boolean writeToBufferIsSlow, boolean anisotropyHasKnownIssues, boolean isExplicitDepthRequired, boolean multiDrawIndirectHasKnownIssues
-) {
+public record HintsAndWorkarounds(boolean writeToBufferIsSlow, boolean anisotropyHasKnownIssues) {
 }
