--- a/com/mojang/blaze3d/buffers/Std140Builder.java
+++ b/com/mojang/blaze3d/buffers/Std140Builder.java
@@ -86,13 +_,14 @@
         this.buffer.putFloat(x);
         this.buffer.putFloat(y);
         this.buffer.putFloat(z);
+        this.buffer.position(this.buffer.position() + 4);
         return this;
     }
 
     public Std140Builder putVec3(Vector3fc vec) {
         this.align(16);
         vec.get(this.buffer);
-        this.buffer.position(this.buffer.position() + 12);
+        this.buffer.position(this.buffer.position() + 16);
         return this;
     }
 
@@ -101,13 +_,14 @@
         this.buffer.putInt(x);
         this.buffer.putInt(y);
         this.buffer.putInt(z);
+        this.buffer.position(this.buffer.position() + 4);
         return this;
     }
 
     public Std140Builder putIVec3(Vector3ic vec) {
         this.align(16);
         vec.get(this.buffer);
-        this.buffer.position(this.buffer.position() + 12);
+        this.buffer.position(this.buffer.position() + 16);
         return this;
     }
 
