--- a/com/mojang/blaze3d/buffers/Std140SizeCalculator.java
+++ b/com/mojang/blaze3d/buffers/Std140SizeCalculator.java
@@ -43,13 +_,13 @@
 
     public Std140SizeCalculator putVec3() {
         this.align(16);
-        this.size += 12;
+        this.size += 16;
         return this;
     }
 
     public Std140SizeCalculator putIVec3() {
         this.align(16);
-        this.size += 12;
+        this.size += 16;
         return this;
     }
 
