--- a/net/minecraft/client/renderer/feature/phase/TranslucentFeatureRenderPhase.java
+++ b/net/minecraft/client/renderer/feature/phase/TranslucentFeatureRenderPhase.java
@@ -40,10 +_,7 @@
             indices[i] = i++;
         }
 
-        IntArrays.unstableSort(indices, (i1, i2) -> {
-            int byDistance = Floats.compare(this.distances.getFloat(i2), this.distances.getFloat(i1));
-            return byDistance != 0 ? byDistance : Integer.compare(i1, i2);
-        });
+        IntArrays.unstableSort(indices, (i1, i2) -> Floats.compare(this.distances.getFloat(i2), this.distances.getFloat(i1)));
         return indices;
     }
 
