--- a/net/minecraft/world/entity/animal/horse/Donkey.java
+++ b/net/minecraft/world/entity/animal/horse/Donkey.java
@@ -17,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.DONKEY_AMBIENT;
     }
 
@@ -27,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.DONKEY_DEATH;
     }
 
@@ -38,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_30682_) {
+    public SoundEvent getHurtSound(DamageSource p_30682_) {
         return SoundEvents.DONKEY_HURT;
     }
 
