--- a/net/minecraft/world/entity/animal/horse/Horse.java
+++ b/net/minecraft/world/entity/animal/horse/Horse.java
@@ -127,12 +_,12 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.HORSE_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.HORSE_DEATH;
     }
 
@@ -143,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_30720_) {
+    public SoundEvent getHurtSound(DamageSource p_30720_) {
         return SoundEvents.HORSE_HURT;
     }
 
