--- a/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/net/minecraft/world/entity/animal/horse/Llama.java
@@ -271,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.LLAMA_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_30803_) {
+    public SoundEvent getHurtSound(DamageSource p_30803_) {
         return SoundEvents.LLAMA_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.LLAMA_DEATH;
     }
 
