--- a/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/net/minecraft/world/entity/animal/horse/Llama.java
@@ -74,16 +_,23 @@
     @Nullable
     private Llama caravanHead;
     @Nullable
-    private Llama caravanTail;
+    public Llama caravanTail;
 
     public Llama(EntityType<? extends Llama> p_30750_, Level p_30751_) {
         super(p_30750_, p_30751_);
+        this.maxDomestication = 30; // Paper - Missing entity API; configure max temper instead of a hardcoded value
     }
 
     public boolean isTraderLlama() {
         return false;
     }
 
+    // CraftBukkit start
+    public void setStrengthPublic(int i) {
+        this.setStrength(i);
+    }
+    // CraftBukkit end
+
     private void setStrength(int p_30841_) {
         this.entityData.set(DATA_STRENGTH_ID, Math.max(1, Math.min(5, p_30841_)));
     }
@@ -309,7 +_,7 @@
 
     @Override
     public int getMaxTemper() {
-        return 30;
+        return super.getMaxTemper(); // Paper - Missing entity API; delegate to parent
     }
 
     @Override
