--- a/net/minecraft/world/entity/animal/horse/Mule.java
+++ b/net/minecraft/world/entity/animal/horse/Mule.java
@@ -16,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.MULE_AMBIENT;
     }
 
@@ -26,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.MULE_DEATH;
     }
 
@@ -37,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_30886_) {
+    public SoundEvent getHurtSound(DamageSource p_30886_) {
         return SoundEvents.MULE_HURT;
     }
 
