--- a/net/minecraft/world/entity/animal/horse/SkeletonTrapGoal.java
+++ b/net/minecraft/world/entity/animal/horse/SkeletonTrapGoal.java
@@ -25,12 +_,16 @@
 
     @Override
     public boolean canUse() {
-        return this.horse.level().hasNearbyAlivePlayer(this.horse.getX(), this.horse.getY(), this.horse.getZ(), 10.0);
+        return this.horse.level().hasNearbyAlivePlayerThatAffectsSpawning(this.horse.getX(), this.horse.getY(), this.horse.getZ(), 10.0); // Paper - Affects Spawning API
     }
 
     @Override
     public void tick() {
         ServerLevel serverlevel = (ServerLevel)this.horse.level();
+        // Forge: Trigger the trap in a tick task to avoid crashes when mods add goals to skeleton horses
+        // (MC-206338/Forge PR #7509)
+        serverlevel.getServer().tell(new net.minecraft.server.TickTask(serverlevel.getServer().getTickCount(), () -> {
+        if (!this.horse.isAlive()) return;
         DifficultyInstance difficultyinstance = serverlevel.getCurrentDifficultyAt(this.horse.blockPosition());
         this.horse.setTrap(false);
         this.horse.setTamed(true);
@@ -39,10 +_,11 @@
         if (lightningbolt != null) {
             lightningbolt.moveTo(this.horse.getX(), this.horse.getY(), this.horse.getZ());
             lightningbolt.setVisualOnly(true);
-            serverlevel.addFreshEntity(lightningbolt);
+            serverlevel.strikeLightning(lightningbolt, org.bukkit.event.weather.LightningStrikeEvent.Cause.TRAP); // CraftBukkit
             Skeleton skeleton = this.createSkeleton(difficultyinstance, this.horse);
             if (skeleton != null) {
                 skeleton.startRiding(this.horse);
+                skeleton.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.TRAP); // CraftBukkit
                 serverlevel.addFreshEntityWithPassengers(skeleton);
 
                 for (int i = 0; i < 3; i++) {
@@ -52,12 +_,14 @@
                         if (skeleton1 != null) {
                             skeleton1.startRiding(abstracthorse);
                             abstracthorse.push(this.horse.getRandom().triangle(0.0, 1.1485), 0.0, this.horse.getRandom().triangle(0.0, 1.1485));
+                            abstracthorse.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.JOCKEY); // CraftBukkit
                             serverlevel.addFreshEntityWithPassengers(abstracthorse);
                         }
                     }
                 }
             }
         }
+        }));
     }
 
     @Nullable
