--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.item;
 
+import com.mohistmc.youer.feature.ban.bans.BanItem;
 import java.util.Objects;
 import java.util.UUID;
 import net.minecraft.core.BlockPos;
@@ -20,7 +_,6 @@
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EntityTypes;
 import net.minecraft.world.entity.Mob;
-import net.minecraft.world.entity.MoveSimulationType;
 import net.minecraft.world.entity.MoverType;
 import net.minecraft.world.entity.SlotAccess;
 import net.minecraft.world.entity.TraceableEntity;
@@ -37,6 +_,14 @@
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityPickupItemEvent;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.player.PlayerPickupItemEvent;
+// CraftBukkit end
+
 public class ItemEntity extends Entity implements TraceableEntity {
     private static final EntityDataAccessor<ItemStack> DATA_ITEM = SynchedEntityData.defineId(ItemEntity.class, EntityDataSerializers.ITEM_STACK);
     private static final float FLOAT_HEIGHT = 0.1F;
@@ -53,6 +_,20 @@
     public @Nullable EntityReference<Entity> thrower;
     public @Nullable UUID target;
     public final float bobOffs = this.random.nextFloat() * (float) Math.PI * 2.0F;
+    /**
+     * The maximum age of this EntityItem. The item is expired once this is reached.
+     */
+    public int lifespan = ItemEntity.LIFETIME;
+
+    public boolean canMobPickup = true; // Paper - Item#canEntityPickup
+    private int despawnRate = -1; // Paper - Alternative item-despawn-rate
+    public net.kyori.adventure.util.TriState frictionState = net.kyori.adventure.util.TriState.NOT_SET; // Paper - Friction API
+    // Purpur start - Item entity immunities
+    public boolean immuneToCactus = false;
+    public boolean immuneToExplosion = false;
+    public boolean immuneToFire = false;
+    public boolean immuneToLightning = false;
+    // Purpur end - Item entity immunities
 
     public ItemEntity(EntityType<? extends ItemEntity> type, Level level) {
         super(type, level);
@@ -64,6 +_,7 @@
         this.setPos(x, y, z);
         this.setItem(itemStack);
         this.setDeltaMovement(this.random.nextDouble() * 0.2 - 0.1, 0.2, this.random.nextDouble() * 0.2 - 0.1);
+        this.lifespan = (itemStack.getItem() == null ? LIFETIME : itemStack.getEntityLifespan(level));
     }
 
     public ItemEntity(Level level, double x, double y, double z, ItemStack itemStack, double deltaX, double deltaY, double deltaZ) {
@@ -71,7 +_,16 @@
         this.setPos(x, y, z);
         this.setItem(itemStack);
         this.setDeltaMovement(deltaX, deltaY, deltaZ);
-    }
+        this.lifespan = (itemStack.getItem() == null ? LIFETIME : itemStack.getEntityLifespan(level));
+    }
+
+    // Paper start - Require item entities to send their location precisely (Fixes MC-4)
+    {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().misc.sendFullPosForItemEntities) {
+            this.setRequiresPrecisePosition(true);
+        }
+    }
+    // Paper end - Require item entities to send their location precisely (Fixes MC-4)
 
     @Override
     public boolean dampensVibrations() {
@@ -106,20 +_,56 @@
         return 0.04;
     }
 
+    // Paper start - EAR 2
+    @Override
+    public void inactiveTick() {
+        super.inactiveTick();
+        if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
+            this.pickupDelay--;
+        }
+        if (this.age != -32768) {
+            this.age++;
+        }
+
+        if (!this.level().isClientSide() && this.age >= this.despawnRate) {// Paper - Alternative item-despawn-rate
+            // CraftBukkit start - fire ItemDespawnEvent
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callItemDespawnEvent(this).isCancelled()) {
+                this.age = 0;
+                return;
+            }
+            // CraftBukkit end
+            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+        }
+    }
+    // Paper end - EAR 2
+
+    // DivineMC start - Lag compensation
+    private void lagCompensation() {
+        if ((this).level().isClientSide()) return;
+
+        if (pickupDelay == 0) return;
+
+        if (pickupDelay - ((ServerLevel) this.level()).tpsCalculator.applicableMissedTicks() <= 0) {
+            pickupDelay = 0;
+            return;
+        }
+
+        pickupDelay = pickupDelay - ((ServerLevel) this.level()).tpsCalculator.applicableMissedTicks();
+    }
+    // DivineMC end - Lag compensation
+
     @Override
     public void tick() {
+        lagCompensation(); // DivineMC - Lag compensation
+        if (getItem().onEntityItemUpdate(this)) return;
         if (this.getItem().isEmpty()) {
-            this.discard();
+            this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
         } else {
             super.tick();
             if (this.pickupDelay > 0 && this.pickupDelay != 32767) {
                 this.pickupDelay--;
             }
 
-            if (this.invulnerableTime > 0) {
-                this.invulnerableTime--;
-            }
-
             this.xo = this.getX();
             this.yo = this.getY();
             this.zo = this.getZ();
@@ -148,8 +_,12 @@
                 this.applyEffectsFromBlocks();
                 float airDrag = this.getAirDrag();
                 float groundFriction = airDrag;
-                if (this.onGround()) {
-                    groundFriction *= this.level().getBlockState(this.getBlockPosBelowThatAffectsMyMovement()).getBlock().getFriction();
+                if (this.frictionState == net.kyori.adventure.util.TriState.FALSE) { // Paper - Friction API
+                    groundFriction = 1F;
+                } else if (this.onGround()) {
+                // Paper end - Friction API
+                    BlockPos groundPos = getBlockPosBelowThatAffectsMyMovement();
+                    groundFriction *= this.level().getBlockState(groundPos).getFriction(level(), groundPos, this);
                 }
 
                 this.setDeltaMovement(this.getDeltaMovement().multiply(groundFriction, airDrag, groundFriction));
@@ -169,9 +_,11 @@
                 this.mergeWithNeighbours();
             }
 
+            /* CraftBukkit start - moved up
             if (this.age != -32768) {
                 this.age++;
             }
+            // CraftBukkit end */
 
             this.needsSync = this.needsSync | this.updateFluidInteraction();
             if (!this.level().isClientSide()) {
@@ -181,8 +_,22 @@
                 }
             }
 
-            if (!this.level().isClientSide() && this.age >= 6000) {
-                this.discard();
+            ItemStack item = this.getItem();
+            if (!this.level().isClientSide() && this.age >= lifespan) {
+            	  // Clamping to MAX_VALUE -1 as age is a Short and going above that would produce an infinite lifespan implicitly (accidentally)
+                this.lifespan = Mth.clamp(lifespan + net.neoforged.neoforge.event.EventHooks.onItemExpire(this), 0, Short.MAX_VALUE - 1);
+                if (this.age >= lifespan) {
+                    // CraftBukkit start - fire ItemDespawnEvent
+                    if (CraftEventFactory.callItemDespawnEvent(this).isCancelled()) {
+                        this.age = 0;
+                    } else {
+                        this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                    }
+                    // CraftBukkit end
+                }
+            }
+            if (item.isEmpty() && !this.isRemoved()) {
+                this.discard(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             }
         }
     }
@@ -207,9 +_,20 @@
 
     private void mergeWithNeighbours() {
         if (this.isMergable()) {
+            // Spigot start
+            double radius = this.level().spigotConfig.itemMerge;
             for (ItemEntity entity : this.level()
-                .getEntitiesOfClass(ItemEntity.class, this.getBoundingBox().inflate(0.5, 0.0, 0.5), other -> other != this && other.isMergable())) {
+                .getEntitiesOfClass(ItemEntity.class, this.getBoundingBox().inflate(radius, this.level().paperConfig().entities.behavior.onlyMergeItemsHorizontally ? 0 : radius - 0.5D, radius), other -> other != this && other.isMergable())) { // Spigot // Paper - configuration to only merge items horizontally
+                // Spigot end
                 if (entity.isMergable()) {
+                    // Paper start - Fix items merging through walls
+                    if (this.level().paperConfig().fixes.fixItemsMergingThroughWalls) {
+                        if (this.level().clipDirect(this.position(), entity.position(),
+                            net.minecraft.world.phys.shapes.CollisionContext.of(this)) == net.minecraft.world.phys.HitResult.Type.BLOCK) {
+                            continue;
+                        }
+                    }
+                    // Paper end - Fix items merging through walls
                     this.tryToMerge(entity);
                     if (this.isRemoved()) {
                         break;
@@ -221,14 +_,14 @@
 
     private boolean isMergable() {
         ItemStack item = this.getItem();
-        return this.isAlive() && this.pickupDelay != 32767 && this.age != -32768 && this.age < 6000 && item.getCount() < item.getMaxStackSize();
+        return this.isAlive() && this.pickupDelay != 32767 && this.age != -32768 && this.age < this.despawnRate && item.getCount() < item.getMaxStackSize(); // Paper - Alternative item-despawn-rate
     }
 
     private void tryToMerge(ItemEntity other) {
         ItemStack thisItemStack = this.getItem();
         ItemStack otherItemStack = other.getItem();
         if (Objects.equals(this.target, other.target) && areMergable(thisItemStack, otherItemStack)) {
-            if (otherItemStack.getCount() < thisItemStack.getCount()) {
+            if (true || otherItemStack.getCount() < thisItemStack.getCount()) { // Spigot
                 merge(this, thisItemStack, other, otherItemStack);
             } else {
                 merge(other, otherItemStack, this, thisItemStack);
@@ -255,11 +_,16 @@
     }
 
     private static void merge(ItemEntity toItem, ItemStack toStack, ItemEntity fromItem, ItemStack fromStack) {
+        // CraftBukkit start
+        if (!CraftEventFactory.callItemMergeEvent(fromItem, toItem)) {
+            return;
+        }
+        // CraftBukkit end
         merge(toItem, toStack, fromStack);
         toItem.pickupDelay = Math.max(toItem.pickupDelay, fromItem.pickupDelay);
         toItem.age = Math.min(toItem.age, fromItem.age);
         if (fromStack.isEmpty()) {
-            fromItem.discard();
+            fromItem.discard(EntityRemoveEvent.Cause.MERGE); // CraftBukkit - add Bukkit remove cause);
         }
     }
 
@@ -275,31 +_,41 @@
 
     @Override
     public final boolean hurtClient(DamageSource source) {
-        return this.invulnerableTime <= 0 && !this.isInvulnerableToBase(source) ? this.getItem().canBeHurtBy(source) : false;
+        return this.isInvulnerableToBase(source) ? false : this.getItem().canBeHurtBy(source);
     }
 
     @Override
     public final boolean hurtServer(ServerLevel level, DamageSource source, float damage) {
-        if (this.invulnerableTime <= 0 && !this.isInvulnerableToBase(source)) {
-            if (!level.getGameRules().get(GameRules.MOB_GRIEFING) && source.getEntity() instanceof Mob) {
-                return false;
-            }
-
-            if (!this.getItem().canBeHurtBy(source)) {
-                return false;
-            }
-
+        // Purpur start - Item entity immunities
+        if (
+                (immuneToCactus && source.is(net.minecraft.world.damagesource.DamageTypes.CACTUS)) ||
+                        (immuneToFire && (source.is(net.minecraft.tags.DamageTypeTags.IS_FIRE) || source.is(net.minecraft.world.damagesource.DamageTypes.ON_FIRE) || source.is(net.minecraft.world.damagesource.DamageTypes.IN_FIRE))) ||
+                        (immuneToLightning && source.is(net.minecraft.world.damagesource.DamageTypes.LIGHTNING_BOLT)) ||
+                        (immuneToExplosion && source.is(net.minecraft.tags.DamageTypeTags.IS_EXPLOSION))
+        ) {
+            return false;
+        } else if (this.isInvulnerableToBase(source)) {
+            // Purpur end - Item entity immunities
+            return false;
+        } else if (!level.getGameRules().get(GameRules.MOB_GRIEFING) && source.getEntity() instanceof Mob) {
+            return false;
+        } else if (!this.getItem().canBeHurtBy(source)) {
+            return false;
+        } else {
+            // CraftBukkit start
+            if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, damage)) {
+                return false;
+            }
+            // CraftBukkit end
             this.markHurt();
-            this.health = (int)(this.health - damage);
+            this.health = (int)((float) this.health - damage);
             this.gameEvent(GameEvent.ENTITY_DAMAGE, source.getEntity());
             if (this.health <= 0) {
-                this.getItem().onDestroyed(this);
-                this.discard();
+                this.getItem().onDestroyed(this, source);
+                this.discard(EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause
             }
 
             return true;
-        } else {
-            return false;
         }
     }
 
@@ -313,11 +_,17 @@
         output.putShort("Health", (short)this.health);
         output.putShort("Age", (short)this.age);
         output.putShort("PickupDelay", (short)this.pickupDelay);
+        output.putInt("Lifespan", this.lifespan);
         EntityReference.store(this.thrower, output, "Thrower");
         output.storeNullable("Owner", UUIDUtil.CODEC, this.target);
         if (!this.getItem().isEmpty()) {
             output.store("Item", ItemStack.CODEC, this.getItem());
         }
+        // Paper start - Friction API
+        if (this.frictionState != net.kyori.adventure.util.TriState.NOT_SET) {
+            output.putString("Paper.FrictionState", this.frictionState.toString());
+        }
+        // Paper end - Friction API
     }
 
     @Override
@@ -325,11 +_,21 @@
         this.health = input.getShortOr("Health", (short)5);
         this.age = input.getShortOr("Age", (short)0);
         this.pickupDelay = input.getShortOr("PickupDelay", (short)0);
+        this.lifespan = input.getIntOr("Lifespan", ItemEntity.LIFETIME);
         this.target = input.read("Owner", UUIDUtil.CODEC).orElse(null);
         this.thrower = EntityReference.read(input, "Thrower");
         this.setItem(input.read("Item", ItemStack.CODEC).orElse(ItemStack.EMPTY));
+        // Paper start - Friction API
+        input.getString("Paper.FrictionState").ifPresent(frictionState -> {
+            try {
+                this.frictionState = net.kyori.adventure.util.TriState.valueOf(frictionState);
+            } catch (Exception ignored) {
+                com.mojang.logging.LogUtils.getLogger().error("Unknown friction state {} for {}", frictionState, this);
+            }
+        });
+        // Paper end - Friction API
         if (this.getItem().isEmpty()) {
-            this.discard();
+            this.discard(null); // CraftBukkit - add Bukkit remove cause
         }
     }
 
@@ -337,12 +_,92 @@
     public void playerTouch(Player player) {
         if (!this.level().isClientSide()) {
             ItemStack itemStack = this.getItem();
+            if (BanItem.checkMoShou(player, itemStack)) return;
             Item item = itemStack.getItem();
             int orgCount = itemStack.getCount();
-            if (this.pickupDelay == 0 && (this.target == null || this.target.equals(player.getUUID())) && player.getInventory().add(itemStack)) {
+
+            // CraftBukkit start - fire PlayerPickupItemEvent
+            int canHold = player.getInventory().canHold(itemStack);
+            int remaining = orgCount - canHold;
+            boolean flyAtPlayer = false; // Paper
+
+            // Paper start - PlayerAttemptPickupItemEvent
+            if (this.pickupDelay <= 0) {
+                org.bukkit.event.player.PlayerAttemptPickupItemEvent attemptEvent = new org.bukkit.event.player.PlayerAttemptPickupItemEvent((org.bukkit.entity.Player) player.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
+                this.level().getCraftServer().getPluginManager().callEvent(attemptEvent);
+
+                flyAtPlayer = attemptEvent.getFlyAtPlayer();
+                if (attemptEvent.isCancelled()) {
+                    if (flyAtPlayer) {
+                        player.take(this, orgCount);
+                    }
+
+                    return;
+                }
+            }
+
+            if (this.pickupDelay <= 0 && canHold > 0) {
+                itemStack.setCount(canHold);
+                // Call legacy event
+                PlayerPickupItemEvent playerEvent = new PlayerPickupItemEvent((org.bukkit.entity.Player) player.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
+                playerEvent.setCancelled(!playerEvent.getPlayer().getCanPickupItems());
+                this.level().getCraftServer().getPluginManager().callEvent(playerEvent);
+                flyAtPlayer = playerEvent.getFlyAtPlayer(); // Paper
+                if (playerEvent.isCancelled()) {
+                    itemStack.setCount(orgCount); // SPIGOT-5294 - restore count
+                    // Paper start
+                    if (flyAtPlayer) {
+                        player.take(this, orgCount);
+                    }
+                    // Paper end
+                    return;
+                }
+
+                // Call newer event afterwards
+                EntityPickupItemEvent entityEvent = new EntityPickupItemEvent((org.bukkit.entity.Player) player.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
+                entityEvent.setCancelled(!entityEvent.getEntity().getCanPickupItems());
+                this.level().getCraftServer().getPluginManager().callEvent(entityEvent);
+                if (entityEvent.isCancelled()) {
+                    itemStack.setCount(orgCount); // SPIGOT-5294 - restore count
+                    return;
+                }
+
+                // Update the ItemStack if it was changed in the event
+                ItemStack current = this.getItem();
+                if (!itemStack.equals(current)) {
+                    itemStack = current;
+                } else {
+                    itemStack.setCount(canHold + remaining); // = i
+                }
+
+                // Possibly < 0; fix here so we do not have to modify code below
+                this.pickupDelay = 0;
+            } else if (this.pickupDelay == 0) {
+                // ensure that the code below isn't triggered if canHold says we can't pick the items up
+                this.pickupDelay = -1;
+            }
+            // CraftBukkit end
+            // Paper end - PlayerAttemptPickupItemEvent
+
+            // Neo: Fire item pickup pre/post and adjust handling logic to adhere to the event result.
+            var result = net.neoforged.neoforge.event.EventHooks.fireItemPickupPre(this, player).canPickup();
+            if (result.isFalse()) {
+                return;
+            }
+
+            // Make a copy of the original stack for use in ItemEntityPickupEvent.Post
+            ItemStack originalCopy = itemStack.copy();
+            // Subvert the vanilla conditions (pickup delay and target check) if the result is true.
+            if ((result.isTrue() || this.pickupDelay == 0 && (this.target == null || this.target.equals(player.getUUID()))) && player.getInventory().add(itemStack)) {
+                // Fire ItemEntityPickupEvent.Post
+                net.neoforged.neoforge.event.EventHooks.fireItemPickupPost(this, player, originalCopy);
+                // Update `orgCount` to reflect the actual pickup amount. Vanilla is wrong here and always reports the whole stack.
+                orgCount = originalCopy.getCount() - itemStack.getCount();
+
+                if (flyAtPlayer) // Paper - PlayerPickupItemEvent
                 player.take(this, orgCount);
                 if (itemStack.isEmpty()) {
-                    this.discard();
+                    this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
                     itemStack.setCount(orgCount);
                 }
 
@@ -379,12 +_,28 @@
 
     public void setItem(ItemStack itemStack) {
         this.getEntityData().set(DATA_ITEM, itemStack);
+        this.despawnRate = this.level().paperConfig().entities.spawning.altItemDespawnRate.enabled ? this.level().paperConfig().entities.spawning.altItemDespawnRate.items.getOrDefault(itemStack.getItem(), this.level().spigotConfig.itemDespawnRate) : this.level().spigotConfig.itemDespawnRate; // Paper - Alternative item-despawn-rate
+        // Purpur start - Item entity immunities
+        if (level().purpurConfig.itemImmuneToCactus.contains(itemStack.getItem())) immuneToCactus = true;
+        if (level().purpurConfig.itemImmuneToExplosion.contains(itemStack.getItem())) immuneToExplosion = true;
+        if (level().purpurConfig.itemImmuneToFire.contains(itemStack.getItem())) immuneToFire = true;
+        if (level().purpurConfig.itemImmuneToLightning.contains(itemStack.getItem())) immuneToLightning = true;
+        // level end - Item entity immunities
     }
 
     public void setTarget(@Nullable UUID target) {
         this.target = target;
     }
 
+    /**
+     * Neo: Add getter for ItemEntity's {@link ItemEntity#target target}.
+     * @return The {@link ItemEntity#target target} that can pick up this item entity (if {@code null}, anyone can pick it up)
+     */
+    @Nullable
+    public UUID getTarget() {
+        return this.target;
+    }
+
     public void setThrower(Entity thrower) {
         this.thrower = EntityReference.of(thrower);
     }
@@ -413,10 +_,6 @@
         return this.pickupDelay > 0;
     }
 
-    public void setInvulnerableFor(int ticks) {
-        this.invulnerableTime = ticks;
-    }
-
     public void setUnlimitedLifetime() {
         this.age = -32768;
     }
@@ -427,7 +_,7 @@
 
     public void makeFakeItem() {
         this.setNeverPickUp();
-        this.age = 5999;
+        this.age = this.despawnRate - 1; // Spigot // Paper - Alternative item-despawn-rate
     }
 
     public static float getSpin(float ageInTicks, float bobOffset) {
@@ -447,10 +_,5 @@
     @Override
     public @Nullable SlotAccess getSlot(int slot) {
         return slot == 0 ? SlotAccess.of(this::getItem, this::setItem) : super.getSlot(slot);
-    }
-
-    @Override
-    public MoveSimulationType getMoveSimulationType() {
-        return MoveSimulationType.SERVER_AND_CLIENT;
     }
 }
