--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -51,6 +_,7 @@
     public @Nullable EntityReference<LivingEntity> owner;
     private boolean usedPortal;
     public float explosionPower = 4.0F;
+    public boolean isIncendiary = false; // CraftBukkit - add field
 
     public PrimedTnt(EntityType<? extends PrimedTnt> type, Level level) {
         super(type, level);
