--- a/net/minecraft/network/protocol/game/ClientboundLightUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundLightUpdatePacket.java
@@ -1,8 +_,7 @@
 package net.minecraft.network.protocol.game;
 
-import io.netty.buffer.ByteBuf;
 import java.util.BitSet;
-import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
@@ -10,21 +_,32 @@
 import net.minecraft.world.level.lighting.LevelLightEngine;
 import org.jspecify.annotations.Nullable;
 
-public record ClientboundLightUpdatePacket(int x, int z, ClientboundLightUpdatePacketData lightData) implements Packet<ClientGamePacketListener> {
-    public static final StreamCodec<ByteBuf, ClientboundLightUpdatePacket> STREAM_CODEC = StreamCodec.composite(
-        ByteBufCodecs.VAR_INT,
-        ClientboundLightUpdatePacket::x,
-        ByteBufCodecs.VAR_INT,
-        ClientboundLightUpdatePacket::z,
-        ClientboundLightUpdatePacketData.STREAM_CODEC,
-        ClientboundLightUpdatePacket::lightData,
-        ClientboundLightUpdatePacket::new
+public class ClientboundLightUpdatePacket implements Packet<ClientGamePacketListener> {
+    public static final StreamCodec<FriendlyByteBuf, ClientboundLightUpdatePacket> STREAM_CODEC = Packet.codec(
+        ClientboundLightUpdatePacket::write, ClientboundLightUpdatePacket::new
     );
+    private final int x;
+    private final int z;
+    private final ClientboundLightUpdatePacketData lightData;
 
     public ClientboundLightUpdatePacket(
         ChunkPos pos, LevelLightEngine lightEngine, @Nullable BitSet skyChangedLightSectionFilter, @Nullable BitSet blockChangedLightSectionFilter
     ) {
-        this(pos.x(), pos.z(), new ClientboundLightUpdatePacketData(pos, lightEngine, skyChangedLightSectionFilter, blockChangedLightSectionFilter));
+        this.x = pos.x();
+        this.z = pos.z();
+        this.lightData = new ClientboundLightUpdatePacketData(pos, lightEngine, skyChangedLightSectionFilter, blockChangedLightSectionFilter);
+    }
+
+    private ClientboundLightUpdatePacket(FriendlyByteBuf input) {
+        this.x = input.readVarInt();
+        this.z = input.readVarInt();
+        this.lightData = new ClientboundLightUpdatePacketData(input, this.x, this.z);
+    }
+
+    private void write(FriendlyByteBuf output) {
+        output.writeVarInt(this.x);
+        output.writeVarInt(this.z);
+        this.lightData.write(output);
     }
 
     @Override
@@ -34,5 +_,17 @@
 
     public void handle(ClientGamePacketListener listener) {
         listener.handleLightUpdatePacket(this);
+    }
+
+    public int getX() {
+        return this.x;
+    }
+
+    public int getZ() {
+        return this.z;
+    }
+
+    public ClientboundLightUpdatePacketData getLightData() {
+        return this.lightData;
     }
 }
