--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -156,6 +_,7 @@
             CompoundTag compoundtag = p_195692_.getUpdateTag(p_195692_.getLevel().registryAccess());
             BlockPos blockpos = p_195692_.getBlockPos();
             int i = SectionPos.sectionRelative(blockpos.getX()) << 4 | SectionPos.sectionRelative(blockpos.getZ());
+            p_195692_.sanitizeSentNbt(compoundtag); // Paper - Sanitize sent data
             return new ClientboundLevelChunkPacketData.BlockEntityInfo(i, blockpos.getY(), p_195692_.getType(), compoundtag.isEmpty() ? null : compoundtag);
         }
     }
