--- a/net/minecraft/network/protocol/game/ClientboundAnimatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundAnimatePacket.java
@@ -10,9 +_,11 @@
     public static final StreamCodec<FriendlyByteBuf, ClientboundAnimatePacket> STREAM_CODEC = Packet.codec(
         ClientboundAnimatePacket::write, ClientboundAnimatePacket::new
     );
-    public static final int WAKE_UP = 0;
-    public static final int CRITICAL_HIT = 1;
-    public static final int MAGIC_CRITICAL_HIT = 2;
+    public static final int SWING_MAIN_HAND = 0;
+    public static final int WAKE_UP = 2;
+    public static final int SWING_OFF_HAND = 3;
+    public static final int CRITICAL_HIT = 4;
+    public static final int MAGIC_CRITICAL_HIT = 5;
     private final int id;
     private final int action;
 
