--- a/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
@@ -38,9 +_,11 @@
         this(
             entity.getId(),
             entity.getUUID(),
-            serverEntity.getPositionBase().x(),
-            serverEntity.getPositionBase().y(),
-            serverEntity.getPositionBase().z(),
+            // Paper start - fix entity tracker desync
+            entity.trackingPosition().x(),
+            entity.trackingPosition().y(),
+            entity.trackingPosition().z(),
+            // Paper end - fix entity tracker desync
             serverEntity.getLastSentXRot(),
             serverEntity.getLastSentYRot(),
             entity.getType(),
