--- a/net/minecraft/network/protocol/game/ClientboundLightUpdatePacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLightUpdatePacketData.java
@@ -1,10 +_,11 @@
 package net.minecraft.network.protocol.game;
 
+import com.google.common.collect.Lists;
 import io.netty.buffer.ByteBuf;
-import java.util.ArrayList;
 import java.util.BitSet;
 import java.util.List;
 import net.minecraft.core.SectionPos;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.world.level.ChunkPos;
@@ -13,30 +_,24 @@
 import net.minecraft.world.level.lighting.LevelLightEngine;
 import org.jspecify.annotations.Nullable;
 
-public record ClientboundLightUpdatePacketData(
-    BitSet skyYMask, BitSet blockYMask, BitSet emptySkyYMask, BitSet emptyBlockYMask, List<byte[]> skyUpdates, List<byte[]> blockUpdates
-) {
+public class ClientboundLightUpdatePacketData {
     private static final StreamCodec<ByteBuf, byte[]> DATA_LAYER_STREAM_CODEC = ByteBufCodecs.byteArray(2048);
-    public static final StreamCodec<ByteBuf, ClientboundLightUpdatePacketData> STREAM_CODEC = StreamCodec.composite(
-        ByteBufCodecs.BIT_SET,
-        ClientboundLightUpdatePacketData::skyYMask,
-        ByteBufCodecs.BIT_SET,
-        ClientboundLightUpdatePacketData::blockYMask,
-        ByteBufCodecs.BIT_SET,
-        ClientboundLightUpdatePacketData::emptySkyYMask,
-        ByteBufCodecs.BIT_SET,
-        ClientboundLightUpdatePacketData::emptyBlockYMask,
-        DATA_LAYER_STREAM_CODEC.apply(ByteBufCodecs.list()),
-        ClientboundLightUpdatePacketData::skyUpdates,
-        DATA_LAYER_STREAM_CODEC.apply(ByteBufCodecs.list()),
-        ClientboundLightUpdatePacketData::blockUpdates,
-        ClientboundLightUpdatePacketData::new
-    );
+    private final BitSet skyYMask;
+    private final BitSet blockYMask;
+    private final BitSet emptySkyYMask;
+    private final BitSet emptyBlockYMask;
+    private final List<byte[]> skyUpdates;
+    private final List<byte[]> blockUpdates;
 
     public ClientboundLightUpdatePacketData(
         ChunkPos chunkPos, LevelLightEngine lightEngine, @Nullable BitSet skyChangedLightSectionFilter, @Nullable BitSet blockChangedLightSectionFilter
     ) {
-        this(new BitSet(), new BitSet(), new BitSet(), new BitSet(), new ArrayList<>(), new ArrayList<>());
+        this.skyYMask = new BitSet();
+        this.blockYMask = new BitSet();
+        this.emptySkyYMask = new BitSet();
+        this.emptyBlockYMask = new BitSet();
+        this.skyUpdates = Lists.newArrayList();
+        this.blockUpdates = Lists.newArrayList();
 
         for (int sectionIndex = 0; sectionIndex < lightEngine.getLightSectionCount(); sectionIndex++) {
             if (skyChangedLightSectionFilter == null || skyChangedLightSectionFilter.get(sectionIndex)) {
@@ -49,6 +_,24 @@
         }
     }
 
+    public ClientboundLightUpdatePacketData(FriendlyByteBuf input, int x, int z) {
+        this.skyYMask = input.readBitSet();
+        this.blockYMask = input.readBitSet();
+        this.emptySkyYMask = input.readBitSet();
+        this.emptyBlockYMask = input.readBitSet();
+        this.skyUpdates = input.readList(DATA_LAYER_STREAM_CODEC);
+        this.blockUpdates = input.readList(DATA_LAYER_STREAM_CODEC);
+    }
+
+    public void write(FriendlyByteBuf output) {
+        output.writeBitSet(this.skyYMask);
+        output.writeBitSet(this.blockYMask);
+        output.writeBitSet(this.emptySkyYMask);
+        output.writeBitSet(this.emptyBlockYMask);
+        output.writeCollection(this.skyUpdates, DATA_LAYER_STREAM_CODEC);
+        output.writeCollection(this.blockUpdates, DATA_LAYER_STREAM_CODEC);
+    }
+
     private void prepareSectionData(
         ChunkPos pos, LevelLightEngine lightEngine, LightLayer layer, int sectionIndex, BitSet mask, BitSet emptyMask, List<byte[]> updates
     ) {
@@ -61,5 +_,29 @@
                 updates.add(data.copy().getData());
             }
         }
+    }
+
+    public BitSet getSkyYMask() {
+        return this.skyYMask;
+    }
+
+    public BitSet getEmptySkyYMask() {
+        return this.emptySkyYMask;
+    }
+
+    public List<byte[]> getSkyUpdates() {
+        return this.skyUpdates;
+    }
+
+    public BitSet getBlockYMask() {
+        return this.blockYMask;
+    }
+
+    public BitSet getEmptyBlockYMask() {
+        return this.emptyBlockYMask;
+    }
+
+    public List<byte[]> getBlockUpdates() {
+        return this.blockUpdates;
     }
 }
