--- a/net/minecraft/network/protocol/game/ClientboundSetDefaultSpawnPositionPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSetDefaultSpawnPositionPacket.java
@@ -10,7 +_,7 @@
     public static final StreamCodec<FriendlyByteBuf, ClientboundSetDefaultSpawnPositionPacket> STREAM_CODEC = Packet.codec(
         ClientboundSetDefaultSpawnPositionPacket::write, ClientboundSetDefaultSpawnPositionPacket::new
     );
-    public final BlockPos pos;
+    private final BlockPos pos;
     private final float angle;
 
     public ClientboundSetDefaultSpawnPositionPacket(BlockPos p_133115_, float p_133116_) {
