--- a/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
@@ -30,12 +_,20 @@
 
         for (short packedPos : changes) {
             this.positions[i] = packedPos;
-            this.states[i] = section.getBlockState(
+            this.states[i] = (section != null) ? section.getBlockState(
                 SectionPos.sectionRelativeX(packedPos), SectionPos.sectionRelativeY(packedPos), SectionPos.sectionRelativeZ(packedPos)
-            );
+            ) : net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(); // CraftBukkit - SPIGOT-6076, Mojang bug when empty chunk section notified
             i++;
         }
     }
+
+    // CraftBukkit start - Add constructor
+    public ClientboundSectionBlocksUpdatePacket(SectionPos sectionpos, ShortSet shortset, BlockState[] states) {
+        this.sectionPos = sectionpos;
+        this.positions = shortset.toShortArray();
+        this.states = states;
+    }
+    // CraftBukkit end
 
     private ClientboundSectionBlocksUpdatePacket(FriendlyByteBuf input) {
         this.sectionPos = SectionPos.STREAM_CODEC.decode(input);
