--- a/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
@@ -7,6 +_,7 @@
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
+import org.bukkit.craftbukkit.util.CraftChatMessage;
 
 public record ClientboundSystemChatPacket(Component content, boolean overlay) implements Packet<ClientGamePacketListener> {
     public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundSystemChatPacket> STREAM_CODEC = StreamCodec.composite(
@@ -16,6 +_,17 @@
         ClientboundSystemChatPacket::overlay,
         ClientboundSystemChatPacket::new
     );
+
+    // Spigot start
+    public ClientboundSystemChatPacket(net.md_5.bungee.api.chat.BaseComponent[] content, boolean overlay) {
+        this(CraftChatMessage.fromJSON(net.md_5.bungee.chat.ComponentSerializer.toString(content)), overlay);
+    }
+    // Spigot end
+    // Paper start
+    public ClientboundSystemChatPacket(net.kyori.adventure.text.Component content, boolean overlay) {
+        this(io.papermc.paper.adventure.PaperAdventure.asVanilla(content), overlay);
+    }
+    // Paper end
 
     @Override
     public PacketType<ClientboundSystemChatPacket> type() {
