--- a/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
@@ -37,6 +_,14 @@
         }
     }
 
+    // CraftBukkit start - Add constructor
+    public ClientboundSectionBlocksUpdatePacket(SectionPos sectionpos, ShortSet shortset, BlockState[] states) {
+        this.sectionPos = sectionpos;
+        this.positions = shortset.toShortArray();
+        this.states = states;
+    }
+    // CraftBukkit end
+
     private ClientboundSectionBlocksUpdatePacket(FriendlyByteBuf input) {
         this.sectionPos = SectionPos.STREAM_CODEC.decode(input);
         int count = input.readVarInt();
