--- a/net/minecraft/network/protocol/game/ClientboundSetPlayerTeamPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSetPlayerTeamPacket.java
@@ -58,6 +_,11 @@
         );
     }
 
+    // Paper start - Multiple Entries with Scoreboards
+    public static ClientboundSetPlayerTeamPacket createMultiplePlayerPacket(PlayerTeam team, Collection<String> players, ClientboundSetPlayerTeamPacket.Action operation) {
+        return new ClientboundSetPlayerTeamPacket(team.getName(), operation == ClientboundSetPlayerTeamPacket.Action.ADD ? 3 : 4, Optional.empty(), players);
+    }
+    // Paper end - Multiple Entries with Scoreboards
     private ClientboundSetPlayerTeamPacket(RegistryFriendlyByteBuf p_323831_) {
         this.name = p_323831_.readUtf();
         this.method = p_323831_.readByte();
@@ -199,7 +_,7 @@
             ComponentSerialization.TRUSTED_STREAM_CODEC.encode(p_323527_, this.displayName);
             p_323527_.writeByte(this.options);
             p_323527_.writeUtf(this.nametagVisibility);
-            p_323527_.writeUtf(this.collisionRule);
+            p_323527_.writeUtf(!io.papermc.paper.configuration.GlobalConfiguration.get().collisions.enablePlayerCollisions ? "never" : this.collisionRule); // Paper - Configurable player collision
             p_323527_.writeEnum(this.color);
             ComponentSerialization.TRUSTED_STREAM_CODEC.encode(p_323527_, this.playerPrefix);
             ComponentSerialization.TRUSTED_STREAM_CODEC.encode(p_323527_, this.playerSuffix);
