--- a/net/minecraft/network/protocol/game/ClientboundRemoveEntitiesPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundRemoveEntitiesPacket.java
@@ -1,21 +_,32 @@
 package net.minecraft.network.protocol.game;
 
-import io.netty.buffer.ByteBuf;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 import it.unimi.dsi.fastutil.ints.IntList;
-import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
 
-public record ClientboundRemoveEntitiesPacket(IntList entityIds) implements Packet<ClientGamePacketListener> {
-    private static final StreamCodec<ByteBuf, IntList> ID_LIST_STREAM_CODEC = ByteBufCodecs.VAR_INT.apply(ByteBufCodecs.collection(IntArrayList::new));
-    public static final StreamCodec<ByteBuf, ClientboundRemoveEntitiesPacket> STREAM_CODEC = StreamCodec.composite(
-        ID_LIST_STREAM_CODEC, ClientboundRemoveEntitiesPacket::entityIds, ClientboundRemoveEntitiesPacket::new
+public class ClientboundRemoveEntitiesPacket implements Packet<ClientGamePacketListener> {
+    public static final StreamCodec<FriendlyByteBuf, ClientboundRemoveEntitiesPacket> STREAM_CODEC = Packet.codec(
+        ClientboundRemoveEntitiesPacket::write, ClientboundRemoveEntitiesPacket::new
     );
+    private final IntList entityIds;
+
+    public ClientboundRemoveEntitiesPacket(IntList ids) {
+        this.entityIds = new IntArrayList(ids);
+    }
 
     public ClientboundRemoveEntitiesPacket(int... ids) {
-        this(IntList.of(ids));
+        this.entityIds = new IntArrayList(ids);
+    }
+
+    private ClientboundRemoveEntitiesPacket(FriendlyByteBuf input) {
+        this.entityIds = input.readIntIdList();
+    }
+
+    private void write(FriendlyByteBuf output) {
+        output.writeIntIdList(this.entityIds);
     }
 
     @Override
@@ -25,5 +_,9 @@
 
     public void handle(ClientGamePacketListener listener) {
         listener.handleRemoveEntities(this);
+    }
+
+    public IntList getEntityIds() {
+        return this.entityIds;
     }
 }
