--- a/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundAddEntityPacket.java
@@ -42,9 +_,11 @@
         this(
             p_131481_.getId(),
             p_131481_.getUUID(),
-            p_352292_.getPositionBase().x(),
-            p_352292_.getPositionBase().y(),
-            p_352292_.getPositionBase().z(),
+            // Paper start - fix entity tracker desync
+            p_131481_.trackingPosition().x(),
+            p_131481_.trackingPosition().y(),
+            p_131481_.trackingPosition().z(),
+            // Paper end - fix entity tracker desync
             p_352292_.getLastSentXRot(),
             p_352292_.getLastSentYRot(),
             p_131481_.getType(),
