--- a/net/minecraft/network/protocol/game/ClientGamePacketListener.java
+++ b/net/minecraft/network/protocol/game/ClientGamePacketListener.java
@@ -18,8 +_,6 @@
 
     void handleHurtAnimation(ClientboundHurtAnimationPacket packet);
 
-    void handleSwingAnimation(ClientboundSwingAnimationPacket packet);
-
     void handleAwardStats(ClientboundAwardStatsPacket packet);
 
     void handleRecipeBookAdd(ClientboundRecipeBookAddPacket packet);
