--- a/net/minecraft/network/protocol/game/ClientboundBossEventPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundBossEventPacket.java
@@ -215,7 +_,7 @@
         }
     }
 
-    public static record UpdateNameOperation(Component name) implements ClientboundBossEventPacket.Operation {
+    static record UpdateNameOperation(Component name) implements ClientboundBossEventPacket.Operation {
         private UpdateNameOperation(RegistryFriendlyByteBuf p_323813_) {
             this(ComponentSerialization.TRUSTED_STREAM_CODEC.decode(p_323813_));
         }
@@ -236,7 +_,7 @@
         }
     }
 
-    public static record UpdateProgressOperation(float progress) implements ClientboundBossEventPacket.Operation {
+    static record UpdateProgressOperation(float progress) implements ClientboundBossEventPacket.Operation {
         private UpdateProgressOperation(RegistryFriendlyByteBuf p_324271_) {
             this(p_324271_.readFloat());
         }
@@ -257,12 +_,12 @@
         }
     }
 
-    public static class UpdatePropertiesOperation implements ClientboundBossEventPacket.Operation {
+    static class UpdatePropertiesOperation implements ClientboundBossEventPacket.Operation {
         private final boolean darkenScreen;
         private final boolean playMusic;
         private final boolean createWorldFog;
 
-        public UpdatePropertiesOperation(boolean p_178751_, boolean p_178752_, boolean p_178753_) {
+        UpdatePropertiesOperation(boolean p_178751_, boolean p_178752_, boolean p_178753_) {
             this.darkenScreen = p_178751_;
             this.playMusic = p_178752_;
             this.createWorldFog = p_178753_;
@@ -291,7 +_,7 @@
         }
     }
 
-    public static class UpdateStyleOperation implements ClientboundBossEventPacket.Operation {
+    static class UpdateStyleOperation implements ClientboundBossEventPacket.Operation {
         private final BossEvent.BossBarColor color;
         private final BossEvent.BossBarOverlay overlay;
 
