--- a/net/minecraft/network/protocol/game/ClientboundChunksBiomesPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundChunksBiomesPacket.java
@@ -4,7 +_,6 @@
 import io.netty.buffer.Unpooled;
 import java.util.List;
 import net.minecraft.network.FriendlyByteBuf;
-import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
@@ -13,17 +_,23 @@
 import net.minecraft.world.level.chunk.LevelChunkSection;
 
 public record ClientboundChunksBiomesPacket(List<ClientboundChunksBiomesPacket.ChunkBiomeData> chunkBiomeData) implements Packet<ClientGamePacketListener> {
-    public static final StreamCodec<ByteBuf, ClientboundChunksBiomesPacket> STREAM_CODEC = StreamCodec.composite(
-        ClientboundChunksBiomesPacket.ChunkBiomeData.STREAM_CODEC.apply(ByteBufCodecs.list()),
-        ClientboundChunksBiomesPacket::chunkBiomeData,
-        ClientboundChunksBiomesPacket::new
+    public static final StreamCodec<FriendlyByteBuf, ClientboundChunksBiomesPacket> STREAM_CODEC = Packet.codec(
+        ClientboundChunksBiomesPacket::write, ClientboundChunksBiomesPacket::new
     );
     private static final int TWO_MEGABYTES = 2097152;
 
+    private ClientboundChunksBiomesPacket(FriendlyByteBuf input) {
+        this(input.readList(ClientboundChunksBiomesPacket.ChunkBiomeData::new));
+    }
+
     public static ClientboundChunksBiomesPacket forChunks(List<LevelChunk> chunks) {
         return new ClientboundChunksBiomesPacket(chunks.stream().map(ClientboundChunksBiomesPacket.ChunkBiomeData::new).toList());
     }
 
+    private void write(FriendlyByteBuf output) {
+        output.writeCollection(this.chunkBiomeData, (o, c) -> c.write(o));
+    }
+
     @Override
     public PacketType<ClientboundChunksBiomesPacket> type() {
         return GamePacketTypes.CLIENTBOUND_CHUNKS_BIOMES;
@@ -34,19 +_,15 @@
     }
 
     public record ChunkBiomeData(ChunkPos pos, byte[] buffer) {
-        public static final StreamCodec<ByteBuf, ClientboundChunksBiomesPacket.ChunkBiomeData> STREAM_CODEC = StreamCodec.composite(
-            ChunkPos.STREAM_CODEC,
-            ClientboundChunksBiomesPacket.ChunkBiomeData::pos,
-            ByteBufCodecs.byteArray(2097152),
-            ClientboundChunksBiomesPacket.ChunkBiomeData::buffer,
-            ClientboundChunksBiomesPacket.ChunkBiomeData::new
-        );
-
         public ChunkBiomeData(LevelChunk chunk) {
             this(chunk.getPos(), new byte[calculateChunkSize(chunk)]);
             extractChunkData(new FriendlyByteBuf(this.getWriteBuffer()), chunk);
         }
 
+        public ChunkBiomeData(FriendlyByteBuf input) {
+            this(input.readChunkPos(), input.readByteArray(2097152));
+        }
+
         private static int calculateChunkSize(LevelChunk chunk) {
             int total = 0;
 
@@ -75,6 +_,11 @@
             if (buffer.writerIndex() != buffer.capacity()) {
                 throw new IllegalStateException("Didn't fill biome buffer: expected " + buffer.capacity() + " bytes, got " + buffer.writerIndex());
             }
+        }
+
+        public void write(FriendlyByteBuf output) {
+            output.writeChunkPos(this.pos);
+            output.writeByteArray(this.buffer);
         }
     }
 }
