--- a/net/minecraft/network/protocol/game/ClientboundMoveEntityPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundMoveEntityPacket.java
@@ -9,7 +_,7 @@
 import net.minecraft.world.level.Level;
 
 public abstract class ClientboundMoveEntityPacket implements Packet<ClientGamePacketListener> {
-    protected final int entityId;
+    public final int entityId;
     protected final short xa;
     protected final short ya;
     protected final short za;
