--- a/net/minecraft/network/protocol/game/ClientboundLoginPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundLoginPacket.java
@@ -3,8 +_,8 @@
 import com.google.common.collect.Sets;
 import java.util.Set;
 import net.minecraft.core.registries.Registries;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.RegistryFriendlyByteBuf;
-import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
@@ -25,33 +_,41 @@
     boolean onlineMode,
     boolean enforcesSecureChat
 ) implements Packet<ClientGamePacketListener> {
-    public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundLoginPacket> STREAM_CODEC = StreamCodec.composite(
-        ByteBufCodecs.INT,
-        ClientboundLoginPacket::playerId,
-        ByteBufCodecs.BOOL,
-        ClientboundLoginPacket::hardcore,
-        ResourceKey.streamCodec(Registries.DIMENSION).apply(ByteBufCodecs.collection(Sets::newHashSetWithExpectedSize)),
-        ClientboundLoginPacket::levels,
-        ByteBufCodecs.VAR_INT,
-        ClientboundLoginPacket::maxPlayers,
-        ByteBufCodecs.VAR_INT,
-        ClientboundLoginPacket::chunkRadius,
-        ByteBufCodecs.VAR_INT,
-        ClientboundLoginPacket::simulationDistance,
-        ByteBufCodecs.BOOL,
-        ClientboundLoginPacket::reducedDebugInfo,
-        ByteBufCodecs.BOOL,
-        ClientboundLoginPacket::showDeathScreen,
-        ByteBufCodecs.BOOL,
-        ClientboundLoginPacket::doLimitedCrafting,
-        CommonPlayerSpawnInfo.STREAM_CODEC,
-        ClientboundLoginPacket::commonPlayerSpawnInfo,
-        ByteBufCodecs.BOOL,
-        ClientboundLoginPacket::onlineMode,
-        ByteBufCodecs.BOOL,
-        ClientboundLoginPacket::enforcesSecureChat,
-        ClientboundLoginPacket::new
+    public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundLoginPacket> STREAM_CODEC = Packet.codec(
+        ClientboundLoginPacket::write, ClientboundLoginPacket::new
     );
+
+    private ClientboundLoginPacket(RegistryFriendlyByteBuf input) {
+        this(
+            input.readInt(),
+            input.readBoolean(),
+            input.readCollection(Sets::newHashSetWithExpectedSize, buf -> buf.readResourceKey(Registries.DIMENSION)),
+            input.readVarInt(),
+            input.readVarInt(),
+            input.readVarInt(),
+            input.readBoolean(),
+            input.readBoolean(),
+            input.readBoolean(),
+            new CommonPlayerSpawnInfo(input),
+            input.readBoolean(),
+            input.readBoolean()
+        );
+    }
+
+    private void write(RegistryFriendlyByteBuf output) {
+        output.writeInt(this.playerId);
+        output.writeBoolean(this.hardcore);
+        output.writeCollection(this.levels, FriendlyByteBuf::writeResourceKey);
+        output.writeVarInt(this.maxPlayers);
+        output.writeVarInt(this.chunkRadius);
+        output.writeVarInt(this.simulationDistance);
+        output.writeBoolean(this.reducedDebugInfo);
+        output.writeBoolean(this.showDeathScreen);
+        output.writeBoolean(this.doLimitedCrafting);
+        this.commonPlayerSpawnInfo.write(output);
+        output.writeBoolean(this.onlineMode);
+        output.writeBoolean(this.enforcesSecureChat);
+    }
 
     @Override
     public PacketType<ClientboundLoginPacket> type() {
