--- a/net/minecraft/network/protocol/game/ClientboundPlayerInfoRemovePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundPlayerInfoRemovePacket.java
@@ -1,18 +_,25 @@
 package net.minecraft.network.protocol.game;
 
-import io.netty.buffer.ByteBuf;
 import java.util.List;
 import java.util.UUID;
 import net.minecraft.core.UUIDUtil;
-import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
 
 public record ClientboundPlayerInfoRemovePacket(List<UUID> profileIds) implements Packet<ClientGamePacketListener> {
-    public static final StreamCodec<ByteBuf, ClientboundPlayerInfoRemovePacket> STREAM_CODEC = StreamCodec.composite(
-        UUIDUtil.STREAM_CODEC.apply(ByteBufCodecs.list()), ClientboundPlayerInfoRemovePacket::profileIds, ClientboundPlayerInfoRemovePacket::new
+    public static final StreamCodec<FriendlyByteBuf, ClientboundPlayerInfoRemovePacket> STREAM_CODEC = Packet.codec(
+        ClientboundPlayerInfoRemovePacket::write, ClientboundPlayerInfoRemovePacket::new
     );
+
+    private ClientboundPlayerInfoRemovePacket(FriendlyByteBuf input) {
+        this(input.readList(UUIDUtil.STREAM_CODEC));
+    }
+
+    private void write(FriendlyByteBuf output) {
+        output.writeCollection(this.profileIds, UUIDUtil.STREAM_CODEC);
+    }
 
     @Override
     public PacketType<ClientboundPlayerInfoRemovePacket> type() {
