--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -6,8 +_,8 @@
 import java.util.EnumMap;
 import java.util.List;
 import java.util.Map;
-import java.util.Optional;
 import java.util.Map.Entry;
+import java.util.function.Consumer;
 import java.util.stream.Collectors;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
@@ -28,40 +_,70 @@
     private static final StreamCodec<ByteBuf, Map<Heightmap.Types, long[]>> HEIGHTMAPS_STREAM_CODEC = ByteBufCodecs.map(
         size -> new EnumMap<>(Heightmap.Types.class), Heightmap.Types.STREAM_CODEC, ByteBufCodecs.LONG_ARRAY
     );
-    private static final int TWO_MEGABYTES = 2097152;
-    public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundLevelChunkPacketData> STREAM_CODEC = StreamCodec.composite(
-        HEIGHTMAPS_STREAM_CODEC,
-        p -> p.heightmaps,
-        ByteBufCodecs.byteArray(2097152),
-        p -> p.buffer,
-        ClientboundLevelChunkPacketData.BlockEntityInfo.STREAM_CODEC.apply(ByteBufCodecs.list()),
-        p -> p.blockEntitiesData,
-        ClientboundLevelChunkPacketData::new
-    );
+    private static final int TWO_MEGABYTES = 2097152; // Paper - diff on change - if this changes, update PacketEncoder
     private final Map<Heightmap.Types, long[]> heightmaps;
     private final byte[] buffer;
     private final List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData;
-
-    public ClientboundLevelChunkPacketData(LevelChunk levelChunk) {
+    // Paper start - Handle oversized block entities in chunks
+    private final java.util.List<net.minecraft.network.protocol.Packet<?>> extraPackets = new java.util.ArrayList<>();
+    private static final int BLOCK_ENTITY_LIMIT = Integer.getInteger("Paper.excessiveTELimit", 750);
+
+    public List<net.minecraft.network.protocol.Packet<?>> getExtraPackets() {
+        return this.extraPackets;
+    }
+    // Paper end - Handle oversized block entities in chunks
+
+    @Deprecated @io.papermc.paper.annotation.DoNotUse // Paper - Anti-Xray - Add chunk packet info
+    public ClientboundLevelChunkPacketData(final LevelChunk levelChunk) {
+        // Paper start - Anti-Xray - Add chunk packet info
+        this(levelChunk, null);
+    }
+    public ClientboundLevelChunkPacketData(final LevelChunk levelChunk, final io.papermc.paper.antixray.ChunkPacketInfo<net.minecraft.world.level.block.state.BlockState> chunkPacketInfo) {
+        // Paper end - Anti-Xray - Add chunk packet info
         this.heightmaps = levelChunk.getHeightmaps()
-            .stream()
-            .filter(entryx -> ((Heightmap.Types)entryx.getKey()).sendToClient())
-            .collect(Collectors.toMap(Entry::getKey, entryx -> (long[])((Heightmap)entryx.getValue()).getRawData().clone()));
+                .stream()
+                .filter(entryx -> ((Heightmap.Types)entryx.getKey()).sendToClient())
+                .collect(Collectors.toMap(Entry::getKey, entryx -> (long[])((Heightmap)entryx.getValue()).getRawData().clone()));
         this.buffer = new byte[calculateChunkSize(levelChunk)];
-        extractChunkData(new FriendlyByteBuf(this.getWriteBuffer()), levelChunk);
+        // Paper start - Anti-Xray - Add chunk packet info
+        if (chunkPacketInfo != null) {
+            chunkPacketInfo.setBuffer(this.buffer);
+        }
+        extractChunkData(new FriendlyByteBuf(this.getWriteBuffer()), levelChunk, chunkPacketInfo);
         this.blockEntitiesData = Lists.newArrayList();
+        int totalTileEntities = 0; // Paper - Handle oversized block entities in chunks
 
         for (Entry<BlockPos, BlockEntity> entry : levelChunk.getBlockEntities().entrySet()) {
+            // Paper start - Handle oversized block entities in chunks
+            if (++totalTileEntities > BLOCK_ENTITY_LIMIT) {
+                net.minecraft.network.protocol.Packet<ClientGamePacketListener> packet = entry.getValue().getUpdatePacket();
+                if (packet != null) {
+                    this.extraPackets.add(packet);
+                    continue;
+                }
+            }
+            // Paper end - Handle oversized block entities in chunks
             this.blockEntitiesData.add(ClientboundLevelChunkPacketData.BlockEntityInfo.create(entry.getValue()));
         }
     }
 
-    private ClientboundLevelChunkPacketData(
-        Map<Heightmap.Types, long[]> heightmaps, byte[] buffer, List<ClientboundLevelChunkPacketData.BlockEntityInfo> blockEntitiesData
-    ) {
-        this.heightmaps = heightmaps;
-        this.buffer = buffer;
-        this.blockEntitiesData = blockEntitiesData;
+    public ClientboundLevelChunkPacketData(RegistryFriendlyByteBuf input, int x, int z) {
+        this.heightmaps = HEIGHTMAPS_STREAM_CODEC.decode(input);
+        int size = input.readVarInt();
+        if (size > 2097152) {
+            throw new RuntimeException("Chunk Packet trying to allocate too much memory on read.");
+        }
+
+        this.buffer = new byte[size];
+        input.readBytes(this.buffer);
+        this.blockEntitiesData = ClientboundLevelChunkPacketData.BlockEntityInfo.LIST_STREAM_CODEC.decode(input);
+    }
+
+    public void write(RegistryFriendlyByteBuf output) {
+        HEIGHTMAPS_STREAM_CODEC.encode(output, this.heightmaps);
+        output.writeVarInt(this.buffer.length);
+        output.writeBytes(this.buffer);
+        ClientboundLevelChunkPacketData.BlockEntityInfo.LIST_STREAM_CODEC.encode(output, this.blockEntitiesData);
     }
 
     private static int calculateChunkSize(LevelChunk chunk) {
@@ -80,9 +_,17 @@
         return buffer;
     }
 
+    @Deprecated @io.papermc.paper.annotation.DoNotUse // Paper - Anti-Xray - Add chunk packet info
     public static void extractChunkData(FriendlyByteBuf buffer, LevelChunk chunk) {
+        // Paper start - Anti-Xray - Add chunk packet info
+        extractChunkData(buffer, chunk, null);
+    }
+    public static void extractChunkData(FriendlyByteBuf buffer, LevelChunk chunk, final io.papermc.paper.antixray.ChunkPacketInfo<net.minecraft.world.level.block.state.BlockState> chunkPacketInfo) {
+        int chunkSectionIndex = 0;
         for (LevelChunkSection section : chunk.getSections()) {
-            section.write(buffer);
+            section.write(buffer, chunkPacketInfo, chunkSectionIndex);
+            chunkSectionIndex++;
+            // Paper end - Anti-Xray - Add chunk packet info
         }
 
         if (buffer.writerIndex() != buffer.capacity()) {
@@ -90,16 +_,20 @@
         }
     }
 
-    public void forEachBlockEntityTag(int chunkX, int chunkZ, ClientboundLevelChunkPacketData.BlockEntityTagOutput output) {
-        int baseX = 16 * chunkX;
-        int baseZ = 16 * chunkZ;
+    public Consumer<ClientboundLevelChunkPacketData.BlockEntityTagOutput> getBlockEntitiesTagsConsumer(int x, int z) {
+        return output -> this.getBlockEntitiesTags(output, x, z);
+    }
+
+    private void getBlockEntitiesTags(ClientboundLevelChunkPacketData.BlockEntityTagOutput output, int x, int z) {
+        int baseX = 16 * x;
+        int baseZ = 16 * z;
         BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
 
         for (ClientboundLevelChunkPacketData.BlockEntityInfo data : this.blockEntitiesData) {
             int unpackedX = baseX + SectionPos.sectionRelative(data.packedXZ >> 4);
             int unpackedZ = baseZ + SectionPos.sectionRelative(data.packedXZ);
             pos.set(unpackedX, data.y, unpackedZ);
-            output.accept(pos, data.type, data.tag.orElse(null));
+            output.accept(pos, data.type, data.tag);
         }
     }
 
@@ -111,26 +_,45 @@
         return this.heightmaps;
     }
 
-    private record BlockEntityInfo(byte packedXZ, short y, BlockEntityType<?> type, Optional<CompoundTag> tag) {
-        public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundLevelChunkPacketData.BlockEntityInfo> STREAM_CODEC = StreamCodec.composite(
-            ByteBufCodecs.BYTE,
-            ClientboundLevelChunkPacketData.BlockEntityInfo::packedXZ,
-            ByteBufCodecs.SHORT,
-            ClientboundLevelChunkPacketData.BlockEntityInfo::y,
-            ByteBufCodecs.registry(Registries.BLOCK_ENTITY_TYPE),
-            ClientboundLevelChunkPacketData.BlockEntityInfo::type,
-            ByteBufCodecs.OPTIONAL_COMPOUND_TAG,
-            ClientboundLevelChunkPacketData.BlockEntityInfo::tag,
-            ClientboundLevelChunkPacketData.BlockEntityInfo::new
-        );
+    private static class BlockEntityInfo {
+        public static final StreamCodec<RegistryFriendlyByteBuf, ClientboundLevelChunkPacketData.BlockEntityInfo> STREAM_CODEC = StreamCodec.ofMember(
+            ClientboundLevelChunkPacketData.BlockEntityInfo::write, ClientboundLevelChunkPacketData.BlockEntityInfo::new
+        );
+        public static final StreamCodec<RegistryFriendlyByteBuf, List<ClientboundLevelChunkPacketData.BlockEntityInfo>> LIST_STREAM_CODEC = STREAM_CODEC.apply(
+            ByteBufCodecs.list()
+        );
+        private final int packedXZ;
+        private final int y;
+        private final BlockEntityType<?> type;
+        private final @Nullable CompoundTag tag;
+
+        private BlockEntityInfo(int packedXZ, int y, BlockEntityType<?> type, @Nullable CompoundTag tag) {
+            this.packedXZ = packedXZ;
+            this.y = y;
+            this.type = type;
+            this.tag = tag;
+        }
+
+        private BlockEntityInfo(RegistryFriendlyByteBuf input) {
+            this.packedXZ = input.readByte();
+            this.y = input.readShort();
+            this.type = ByteBufCodecs.registry(Registries.BLOCK_ENTITY_TYPE).decode(input);
+            this.tag = input.readNbt();
+        }
+
+        private void write(RegistryFriendlyByteBuf output) {
+            output.writeByte(this.packedXZ);
+            output.writeShort(this.y);
+            ByteBufCodecs.registry(Registries.BLOCK_ENTITY_TYPE).encode(output, this.type);
+            output.writeNbt(this.tag);
+        }
 
         private static ClientboundLevelChunkPacketData.BlockEntityInfo create(BlockEntity blockEntity) {
             CompoundTag tag = blockEntity.getUpdateTag(blockEntity.getLevel().registryAccess());
             BlockPos pos = blockEntity.getBlockPos();
             int xz = SectionPos.sectionRelative(pos.getX()) << 4 | SectionPos.sectionRelative(pos.getZ());
-            return new ClientboundLevelChunkPacketData.BlockEntityInfo(
-                (byte)xz, (short)pos.getY(), blockEntity.getType(), tag.isEmpty() ? Optional.empty() : Optional.of(tag)
-            );
+            blockEntity.sanitizeSentNbt(tag); // Paper - Sanitize sent data
+            return new ClientboundLevelChunkPacketData.BlockEntityInfo(xz, pos.getY(), blockEntity.getType(), tag.isEmpty() ? null : tag);
         }
     }
 
