--- a/net/minecraft/network/protocol/game/GamePacketTypes.java
+++ b/net/minecraft/network/protocol/game/GamePacketTypes.java
@@ -120,7 +_,6 @@
     public static final PacketType<ClientboundSoundPacket> CLIENTBOUND_SOUND = createClientbound("sound");
     public static final PacketType<ClientboundStartConfigurationPacket> CLIENTBOUND_START_CONFIGURATION = createClientbound("start_configuration");
     public static final PacketType<ClientboundStopSoundPacket> CLIENTBOUND_STOP_SOUND = createClientbound("stop_sound");
-    public static final PacketType<ClientboundSwingAnimationPacket> CLIENTBOUND_SWING_ANIMATION = createClientbound("swing_animation");
     public static final PacketType<ClientboundSystemChatPacket> CLIENTBOUND_SYSTEM_CHAT = createClientbound("system_chat");
     public static final PacketType<ClientboundTabListPacket> CLIENTBOUND_TAB_LIST = createClientbound("tab_list");
     public static final PacketType<ClientboundTagQueryPacket> CLIENTBOUND_TAG_QUERY = createClientbound("tag_query");
@@ -179,7 +_,6 @@
     public static final PacketType<ServerboundPlayerCommandPacket> SERVERBOUND_PLAYER_COMMAND = createServerbound("player_command");
     public static final PacketType<ServerboundPlayerInputPacket> SERVERBOUND_PLAYER_INPUT = createServerbound("player_input");
     public static final PacketType<ServerboundPlayerLoadedPacket> SERVERBOUND_PLAYER_LOADED = createServerbound("player_loaded");
-    public static final PacketType<ServerboundPunchPacket> SERVERBOUND_PUNCH = createServerbound("punch");
     public static final PacketType<ServerboundRecipeBookChangeSettingsPacket> SERVERBOUND_RECIPE_BOOK_CHANGE_SETTINGS = createServerbound(
         "recipe_book_change_settings"
     );
@@ -201,6 +_,7 @@
     );
     public static final PacketType<ServerboundSignUpdatePacket> SERVERBOUND_SIGN_UPDATE = createServerbound("sign_update");
     public static final PacketType<ServerboundSpectatorActionPacket> SERVERBOUND_SPECTATOR_ACTION = createServerbound("spectator_action");
+    public static final PacketType<ServerboundSwingPacket> SERVERBOUND_SWING = createServerbound("swing");
     public static final PacketType<ServerboundTeleportToEntityPacket> SERVERBOUND_TELEPORT_TO_ENTITY = createServerbound("teleport_to_entity");
     public static final PacketType<ServerboundUseItemOnPacket> SERVERBOUND_USE_ITEM_ON = createServerbound("use_item_on");
     public static final PacketType<ServerboundUseItemPacket> SERVERBOUND_USE_ITEM = createServerbound("use_item");
