--- a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
@@ -1,10 +_,9 @@
 package net.minecraft.network.protocol.game;
 
-import io.netty.buffer.ByteBuf;
 import java.time.Instant;
 import net.minecraft.commands.arguments.ArgumentSignatures;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.LastSeenMessages;
-import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
@@ -12,19 +_,21 @@
 public record ServerboundChatCommandSignedPacket(
     String command, Instant timeStamp, long salt, ArgumentSignatures argumentSignatures, LastSeenMessages.Update lastSeenMessages
 ) implements Packet<ServerGamePacketListener> {
-    public static final StreamCodec<ByteBuf, ServerboundChatCommandSignedPacket> STREAM_CODEC = StreamCodec.composite(
-        ByteBufCodecs.STRING_UTF8,
-        ServerboundChatCommandSignedPacket::command,
-        ByteBufCodecs.INSTANT,
-        ServerboundChatCommandSignedPacket::timeStamp,
-        ByteBufCodecs.LONG,
-        ServerboundChatCommandSignedPacket::salt,
-        ArgumentSignatures.STREAM_CODEC,
-        ServerboundChatCommandSignedPacket::argumentSignatures,
-        LastSeenMessages.Update.STREAM_CODEC,
-        ServerboundChatCommandSignedPacket::lastSeenMessages,
-        ServerboundChatCommandSignedPacket::new
+    public static final StreamCodec<FriendlyByteBuf, ServerboundChatCommandSignedPacket> STREAM_CODEC = Packet.codec(
+        ServerboundChatCommandSignedPacket::write, ServerboundChatCommandSignedPacket::new
     );
+
+    private ServerboundChatCommandSignedPacket(FriendlyByteBuf input) {
+        this(input.readUtf(net.minecraft.network.protocol.game.ServerboundChatCommandPacket.MAX_CHAT_PACKET_INPUT_SIZE), input.readInstant(), input.readLong(), new ArgumentSignatures(input), new LastSeenMessages.Update(input)); // Paper - limit chat command inputs
+    }
+
+    private void write(FriendlyByteBuf output) {
+        output.writeUtf(this.command);
+        output.writeInstant(this.timeStamp);
+        output.writeLong(this.salt);
+        this.argumentSignatures.write(output);
+        this.lastSeenMessages.write(output);
+    }
 
     @Override
     public PacketType<ServerboundChatCommandSignedPacket> type() {
