--- a/net/minecraft/network/protocol/game/ServerboundMovePlayerPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundMovePlayerPacket.java
@@ -6,14 +_,14 @@
 import net.minecraft.network.protocol.PacketType;
 
 public abstract class ServerboundMovePlayerPacket implements Packet<ServerGamePacketListener> {
-    public final double x;
-    public final double y;
-    public final double z;
-    public final float yRot;
-    public final float xRot;
+    protected final double x;
+    protected final double y;
+    protected final double z;
+    protected final float yRot;
+    protected final float xRot;
     protected final boolean onGround;
-    public final boolean hasPos;
-    public final boolean hasRot;
+    protected final boolean hasPos;
+    protected final boolean hasRot;
 
     protected ServerboundMovePlayerPacket(
         double p_179675_, double p_179676_, double p_179677_, float p_179678_, float p_179679_, boolean p_179680_, boolean p_179681_, boolean p_179682_
