--- a/net/minecraft/network/protocol/game/ServerboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
@@ -1,30 +_,31 @@
 package net.minecraft.network.protocol.game;
 
 import java.time.Instant;
-import java.util.Optional;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.LastSeenMessages;
 import net.minecraft.network.chat.MessageSignature;
-import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
+import org.jspecify.annotations.Nullable;
 
-public record ServerboundChatPacket(String message, Instant timeStamp, long salt, Optional<MessageSignature> signature, LastSeenMessages.Update lastSeenMessages)
+public record ServerboundChatPacket(String message, Instant timeStamp, long salt, @Nullable MessageSignature signature, LastSeenMessages.Update lastSeenMessages)
     implements Packet<ServerGamePacketListener> {
-    public static final StreamCodec<FriendlyByteBuf, ServerboundChatPacket> STREAM_CODEC = StreamCodec.composite(
-        ByteBufCodecs.stringUtf8(256),
-        ServerboundChatPacket::message,
-        ByteBufCodecs.INSTANT,
-        ServerboundChatPacket::timeStamp,
-        ByteBufCodecs.LONG,
-        ServerboundChatPacket::salt,
-        MessageSignature.STREAM_CODEC.apply(ByteBufCodecs::optional),
-        ServerboundChatPacket::signature,
-        LastSeenMessages.Update.STREAM_CODEC,
-        ServerboundChatPacket::lastSeenMessages,
-        ServerboundChatPacket::new
+    public static final StreamCodec<FriendlyByteBuf, ServerboundChatPacket> STREAM_CODEC = Packet.codec(
+        ServerboundChatPacket::write, ServerboundChatPacket::new
     );
+
+    private ServerboundChatPacket(FriendlyByteBuf input) {
+        this(input.readUtf(256), input.readInstant(), input.readLong(), input.readNullable(MessageSignature::read), new LastSeenMessages.Update(input));
+    }
+
+    private void write(FriendlyByteBuf output) {
+        output.writeUtf(this.message, 256);
+        output.writeInstant(this.timeStamp);
+        output.writeLong(this.salt);
+        output.writeNullable(this.signature, MessageSignature::write);
+        this.lastSeenMessages.write(output);
+    }
 
     @Override
     public PacketType<ServerboundChatPacket> type() {
