--- a/net/minecraft/network/protocol/game/ServerboundMoveVehiclePacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundMoveVehiclePacket.java
@@ -18,7 +_,9 @@
     );
 
     public static ServerboundMoveVehiclePacket fromEntity(Entity entity) {
-        return new ServerboundMoveVehiclePacket(entity.getClientPositionAndRotation(), entity.onGround());
+        return entity.isInterpolating()
+            ? new ServerboundMoveVehiclePacket(entity.getInterpolation().getCurrentPositionAndRotation(), entity.onGround())
+            : new ServerboundMoveVehiclePacket(entity.storePositionAndRotation(), entity.onGround());
     }
 
     @Override
