--- a/net/minecraft/network/protocol/game/ServerboundChatSessionUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatSessionUpdatePacket.java
@@ -1,14 +_,23 @@
 package net.minecraft.network.protocol.game;
 
-import io.netty.buffer.ByteBuf;
+import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.RemoteChatSession;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
 
 public record ServerboundChatSessionUpdatePacket(RemoteChatSession.Data chatSession) implements Packet<ServerGamePacketListener> {
-    public static final StreamCodec<ByteBuf, ServerboundChatSessionUpdatePacket> STREAM_CODEC = RemoteChatSession.Data.STREAM_CODEC
-        .map(ServerboundChatSessionUpdatePacket::new, ServerboundChatSessionUpdatePacket::chatSession);
+    public static final StreamCodec<FriendlyByteBuf, ServerboundChatSessionUpdatePacket> STREAM_CODEC = Packet.codec(
+        ServerboundChatSessionUpdatePacket::write, ServerboundChatSessionUpdatePacket::new
+    );
+
+    private ServerboundChatSessionUpdatePacket(FriendlyByteBuf input) {
+        this(RemoteChatSession.Data.read(input));
+    }
+
+    private void write(FriendlyByteBuf output) {
+        RemoteChatSession.Data.write(output, this.chatSession);
+    }
 
     @Override
     public PacketType<ServerboundChatSessionUpdatePacket> type() {
