--- a/net/minecraft/network/protocol/game/VecDeltaCodec.java
+++ b/net/minecraft/network/protocol/game/VecDeltaCodec.java
@@ -1,15 +_,11 @@
 package net.minecraft.network.protocol.game;
 
 import com.google.common.annotations.VisibleForTesting;
-import java.util.List;
-import net.minecraft.world.entity.PositionPath;
-import net.minecraft.world.entity.PositionStep;
 import net.minecraft.world.phys.Vec3;
-import org.jspecify.annotations.Nullable;
 
 public class VecDeltaCodec {
     private static final double TRUNCATION_STEPS = 4096.0;
-    private Vec3 base = Vec3.ZERO;
+    public Vec3 base = Vec3.ZERO; // Paper
 
     @VisibleForTesting
     static long encode(double input) {
@@ -21,14 +_,6 @@
         return v / 4096.0;
     }
 
-    public static double encodingPrecisionLoss(double d) {
-        return decode(encode(d)) - d;
-    }
-
-    public static boolean isDeltaTooBig(long xa, long ya, long za) {
-        return xa < -32768L || xa > 32767L || ya < -32768L || ya > 32767L || za < -32768L || za > 32767L;
-    }
-
     public Vec3 decode(long xa, long ya, long za) {
         if (xa == 0L && ya == 0L && za == 0L) {
             return this.base;
@@ -38,23 +_,6 @@
         double y = ya == 0L ? this.base.y : decode(encode(this.base.y) + ya);
         double z = za == 0L ? this.base.z : decode(encode(this.base.z) + za);
         return new Vec3(x, y, z);
-    }
-
-    public @Nullable VecDelta tryEncode(PositionPath position) {
-        return switch (position) {
-            case PositionPath.Linear(Vec3 pos) -> this.tryEncode(pos);
-            case PositionPath.Stepped(Vec3 _, List<PositionStep> steps) -> {
-                yield VecDelta.Stepped.tryEncode(this, steps);
-            }
-            default -> throw new MatchException(null, null);
-        };
-    }
-
-    public @Nullable VecDelta tryEncode(Vec3 pos) {
-        long xa = this.encodeX(pos);
-        long ya = this.encodeY(pos);
-        long za = this.encodeZ(pos);
-        return isDeltaTooBig(xa, ya, za) ? null : new VecDelta.Linear((short)xa, (short)ya, (short)za);
     }
 
     public long encodeX(Vec3 pos) {
