--- a/net/minecraft/commands/CommandSource.java
+++ b/net/minecraft/commands/CommandSource.java
@@ -1,10 +_,9 @@
 package net.minecraft.commands;
 
 import net.minecraft.network.chat.Component;
-
-import net.neoforged.neoforge.common.extensions.ICommandSourceExtension;
-
-public interface CommandSource extends ICommandSourceExtension {
+import org.bukkit.craftbukkit.command.NullCommandSender;
+
+public interface CommandSource extends net.neoforged.neoforge.common.extensions.ICommandSourceExtension {
     CommandSource NULL = new CommandSource() {
         @Override
         public void sendSystemMessage(Component message) {
@@ -24,6 +_,13 @@
         public boolean shouldInformAdmins() {
             return false;
         }
+
+        // CraftBukkit start
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+            return NullCommandSender.INSTANCE;
+        }
+        // CraftBukkit end
     };
 
     void sendSystemMessage(Component message);
@@ -36,5 +_,9 @@
 
     default boolean alwaysAccepts() {
         return false;
+    }
+
+    default org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+        return NULL.getBukkitSender( wrapper);
     }
 }
