--- a/net/minecraft/commands/CommandSourceStack.java
+++ b/net/minecraft/commands/CommandSourceStack.java
@@ -9,24 +_,25 @@
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import com.mojang.brigadier.suggestion.Suggestions;
 import com.mojang.brigadier.suggestion.SuggestionsBuilder;
+import com.mojang.brigadier.tree.CommandNode;
 import java.util.Collection;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.BinaryOperator;
-import java.util.function.Predicate;
 import java.util.function.Supplier;
 import java.util.stream.Stream;
 import net.minecraft.ChatFormatting;
+import net.minecraft.advancements.AdvancementHolder;
 import net.minecraft.commands.arguments.EntityAnchorArgument;
 import net.minecraft.commands.execution.TraceCallbacks;
 import net.minecraft.core.HolderLookup;
 import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.registries.Registries;
 import net.minecraft.network.chat.ChatType;
-import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentUtils;
 import net.minecraft.network.chat.OutgoingChatMessage;
@@ -35,7 +_,9 @@
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.server.permissions.Permission;
 import net.minecraft.server.permissions.PermissionSet;
+import net.minecraft.server.permissions.Permissions;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.util.Mth;
 import net.minecraft.util.TaskChainer;
@@ -50,14 +_,15 @@
 
 import net.neoforged.neoforge.common.extensions.ICommandSourceStackExtension;
 
-public class CommandSourceStack implements SharedSuggestionProvider, ExecutionCommandSource<CommandSourceStack>, ICommandSourceStackExtension {
+public class CommandSourceStack implements SharedSuggestionProvider, ExecutionCommandSource<CommandSourceStack>, ICommandSourceStackExtension, io.papermc.paper.command.brigadier.PaperCommandSourceStack { // Paper - Brigadier API
     public static final SimpleCommandExceptionType ERROR_NOT_PLAYER = new SimpleCommandExceptionType(Component.translatable("permissions.requires.player"));
     public static final SimpleCommandExceptionType ERROR_NOT_ENTITY = new SimpleCommandExceptionType(Component.translatable("permissions.requires.entity"));
     public final CommandSource source;
     private final Vec3 worldPosition;
     private final ServerLevel level;
     private final PermissionSet permissions;
-    private final CommandSourceStack.NamesProvider namesProvider;
+    private final String textName;
+    private final Component displayName;
     private final MinecraftServer server;
     private final boolean silent;
     private final @Nullable Entity entity;
@@ -66,26 +_,16 @@
     private final Vec2 rotation;
     private final CommandSigningContext signingContext;
     private final TaskChainer chatMessageChainer;
-
-    public CommandSourceStack(
-        CommandSource source, Vec3 position, Vec2 rotation, ServerLevel level, PermissionSet permissions, MinecraftServer server, Entity entity
-    ) {
-        this(source, position, rotation, level, permissions, CommandSourceStack.NamesProvider.FOR_ENTITY, server, entity);
-    }
-
-    public CommandSourceStack(
-        CommandSource source, Vec3 position, Vec2 rotation, ServerLevel level, PermissionSet permissions, Component name, MinecraftServer server
-    ) {
-        this(source, position, rotation, level, permissions, CommandSourceStack.NamesProvider.constant(name), server, null);
-    }
-
-    private CommandSourceStack(
+    public boolean bypassSelectorPermissions = false; // Paper - add bypass for selector permissions
+
+    public CommandSourceStack(
         CommandSource source,
         Vec3 position,
         Vec2 rotation,
         ServerLevel level,
         PermissionSet permissions,
-        CommandSourceStack.NamesProvider namesProvider,
+        String textName,
+        Component displayName,
         MinecraftServer server,
         @Nullable Entity entity
     ) {
@@ -95,7 +_,8 @@
             rotation,
             level,
             permissions,
-            namesProvider,
+            textName,
+            displayName,
             server,
             entity,
             false,
@@ -112,7 +_,8 @@
         Vec2 rotation,
         ServerLevel level,
         PermissionSet permissions,
-        CommandSourceStack.NamesProvider namesProvider,
+        String textName,
+        Component displayName,
         MinecraftServer server,
         @Nullable Entity entity,
         boolean silent,
@@ -127,7 +_,8 @@
         this.silent = silent;
         this.entity = entity;
         this.permissions = permissions;
-        this.namesProvider = namesProvider;
+        this.textName = textName;
+        this.displayName = displayName;
         this.server = server;
         this.resultCallback = resultCallback;
         this.anchor = anchor;
@@ -145,7 +_,8 @@
                 this.rotation,
                 this.level,
                 this.permissions,
-                this.namesProvider,
+                this.textName,
+                this.displayName,
                 this.server,
                 this.entity,
                 this.silent,
@@ -165,7 +_,8 @@
                 this.rotation,
                 this.level,
                 this.permissions,
-                CommandSourceStack.NamesProvider.FOR_ENTITY,
+                entity.getPlainTextName(),
+                entity.getDisplayName(),
                 this.server,
                 entity,
                 this.silent,
@@ -185,7 +_,8 @@
                 this.rotation,
                 this.level,
                 this.permissions,
-                this.namesProvider,
+                this.textName,
+                this.displayName,
                 this.server,
                 this.entity,
                 this.silent,
@@ -195,6 +_,30 @@
                 this.chatMessageChainer
             );
     }
+
+    // Paper start - Expose 'with' functions from the CommandSourceStack
+    @Override
+    public CommandSourceStack withLocation(org.bukkit.Location location) {
+        return this.getLocation().equals(location)
+                ? this
+                : new CommandSourceStack(
+                this.source,
+                new Vec3(location.x(), location.y(), location.z()),
+                new Vec2(location.getPitch(), location.getYaw()),
+                ((org.bukkit.craftbukkit.CraftWorld) location.getWorld()).getHandle(),
+                this.permissions,
+                this.textName,
+                this.displayName,
+                this.server,
+                this.entity,
+                this.silent,
+                this.resultCallback,
+                this.anchor,
+                this.signingContext,
+                this.chatMessageChainer
+        );
+    }
+    // Paper end - Expose 'with' functions from the CommandSourceStack
 
     public CommandSourceStack withRotation(Vec2 rotation) {
         return this.rotation.equals(rotation)
@@ -205,7 +_,8 @@
                 rotation,
                 this.level,
                 this.permissions,
-                this.namesProvider,
+                this.textName,
+                this.displayName,
                 this.server,
                 this.entity,
                 this.silent,
@@ -225,7 +_,8 @@
                 this.rotation,
                 this.level,
                 this.permissions,
-                this.namesProvider,
+                this.textName,
+                this.displayName,
                 this.server,
                 this.entity,
                 this.silent,
@@ -249,7 +_,8 @@
                 this.rotation,
                 this.level,
                 this.permissions,
-                this.namesProvider,
+                this.textName,
+                this.displayName,
                 this.server,
                 this.entity,
                 true,
@@ -270,7 +_,8 @@
                 this.rotation,
                 this.level,
                 permissions,
-                this.namesProvider,
+                this.textName,
+                this.displayName,
                 this.server,
                 this.entity,
                 this.silent,
@@ -294,7 +_,8 @@
                 this.rotation,
                 this.level,
                 this.permissions,
-                this.namesProvider,
+                this.textName,
+                this.displayName,
                 this.server,
                 this.entity,
                 this.silent,
@@ -318,7 +_,8 @@
             this.rotation,
             level,
             this.permissions,
-            this.namesProvider,
+            this.textName,
+            this.displayName,
             this.server,
             this.entity,
             this.silent,
@@ -353,7 +_,8 @@
                 this.rotation,
                 this.level,
                 this.permissions,
-                this.namesProvider,
+                this.textName,
+                this.displayName,
                 this.server,
                 this.entity,
                 this.silent,
@@ -365,11 +_,11 @@
     }
 
     public Component getDisplayName() {
-        return this.namesProvider.displayName(this.entity);
+        return this.displayName;
     }
 
     public String getTextName() {
-        return this.namesProvider.textName(this.entity);
+        return this.textName;
     }
 
     @Override
@@ -377,6 +_,45 @@
         return this.permissions;
     }
 
+    // Paper start - Fix permission levels for command blocks
+    private boolean forceRespectPermissionLevel() {
+        return this.source == CommandSource.NULL || (this.source instanceof final net.minecraft.world.level.BaseCommandBlock commandBlock && commandBlock.getLevel().paperConfig().commandBlocks.forceFollowPermLevel);
+    }
+    // Paper end - Fix permission levels for command blocks
+
+    // CraftBukkit start
+    public boolean hasPermission(net.minecraft.server.permissions.Permission permission, String bukkitPermission) {
+        // Paper start - Fix permission levels for command blocks
+        final java.util.function.BooleanSupplier hasBukkitPerm = () -> this.source == CommandSource.NULL /*treat NULL as having all bukkit perms*/ || this.getBukkitSender().hasPermission(bukkitPermission); // lazily check bukkit perms to the benefit of custom permission setups
+        // if the server is null, we must check the vanilla perm level system
+        // if ignoreVanillaPermissions is true, we can skip vanilla perms and just run the bukkit perm check
+        //noinspection ConstantValue
+        if (this.getServer() == null || !this.getServer().server.ignoreVanillaPermissions) { // server & level are null for command function loading
+            final boolean hasPermLevel = this.permissions.hasPermission(permission);
+            if (this.forceRespectPermissionLevel()) { // NULL CommandSource and command blocks (if setting is enabled) should always pass the vanilla perm check
+                return hasPermLevel && hasBukkitPerm.getAsBoolean();
+            } else { // otherwise check vanilla perm first then bukkit perm, matching upstream behavior
+                return hasPermLevel || hasBukkitPerm.getAsBoolean();
+            }
+        }
+        return hasBukkitPerm.getAsBoolean();
+        // Paper end - Fix permission levels for command blocks
+    }
+    // CraftBukkit end
+
+    // Purpur start - Gamemode extra permissions
+    public boolean testPermission(net.minecraft.server.permissions.Permission permission, String bukkitPermission) {
+        if (hasPermission(permission, bukkitPermission)) {
+            return true;
+        }
+        net.kyori.adventure.text.Component permissionMessage = getLevel().getServer().server.permissionMessage();
+        if (!permissionMessage.equals(net.kyori.adventure.text.Component.empty())) {
+            sendFailure(io.papermc.paper.adventure.PaperAdventure.asVanilla(permissionMessage.replaceText(net.kyori.adventure.text.TextReplacementConfig.builder().matchLiteral("<permission>").replacement(bukkitPermission).build())));
+        }
+        return false;
+    }
+    // Purpur end - Gamemode extra permissions
+
     public Vec3 getPosition() {
         return this.worldPosition;
     }
@@ -460,6 +_,30 @@
         }
     }
 
+    // Purpur start - Purpur config files
+    public void sendSuccess(@Nullable String message) {
+        sendSuccess(message, false);
+    }
+
+    public void sendSuccess(@Nullable String message, boolean broadcastToOps) {
+        if (message == null) {
+            return;
+        }
+        sendSuccess(net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(message), broadcastToOps);
+    }
+
+    public void sendSuccess(net.kyori.adventure.text.@Nullable Component message) {
+        sendSuccess(message, false);
+    }
+
+    public void sendSuccess(net.kyori.adventure.text.@Nullable Component message, boolean broadcastToOps) {
+        if (message == null) {
+            return;
+        }
+        sendSuccess(() -> io.papermc.paper.adventure.PaperAdventure.asVanilla(message), broadcastToOps);
+    }
+    // Purpur end - Purpur config files
+
     public void sendSuccess(Supplier<Component> messageSupplier, boolean broadcast) {
         boolean shouldSendSystemMessage = this.source.acceptsSuccess() && !this.silent;
         boolean shouldBroadcast = broadcast && this.source.shouldInformAdmins() && !this.silent;
@@ -480,20 +_,25 @@
         GameRules gameRules = this.level.getGameRules();
         if (gameRules.get(GameRules.SEND_COMMAND_FEEDBACK)) {
             for (ServerPlayer player : this.server.getPlayerList().getPlayers()) {
-                if (player.commandSource() != this.source && this.server.getPlayerList().isOp(player.nameAndId())) {
+                if (player.commandSource() != this.source && player.getBukkitEntity().hasPermission("minecraft.admin.command_feedback")) { // CraftBukkit
                     player.sendSystemMessage(broadcast);
                 }
             }
         }
 
-        if (this.source != this.server && gameRules.get(GameRules.LOG_ADMIN_COMMANDS)) {
+        if (this.source != this.server && gameRules.get(GameRules.LOG_ADMIN_COMMANDS) && (!org.spigotmc.SpigotConfig.silentCommandBlocks || !(this.source instanceof net.minecraft.world.level.BaseCommandBlock.CloseableCommandBlockSource))) { // Spigot // Paper - Fix spigot config silentCommandBlocks filtering player commands
             this.server.sendSystemMessage(broadcast);
         }
     }
 
     public void sendFailure(Component message) {
+        // Paper start - Add UnknownCommandEvent
+        this.sendFailure(message, true);
+    }
+    public void sendFailure(final Component message, final boolean withStyle) {
+        // Paper end - Add UnknownCommandEvent
         if (this.source.acceptsFailure() && !this.silent) {
-            this.source.sendSystemMessage(Component.empty().append(message).withStyle(ChatFormatting.RED));
+            this.source.sendSystemMessage(withStyle ? Component.empty().append(message).withStyle(ChatFormatting.RED) : message); // Paper - Add UnknownCommandEvent
         }
     }
 
@@ -504,7 +_,7 @@
 
     @Override
     public Collection<String> getOnlinePlayerNames() {
-        return Lists.newArrayList(this.server.getPlayerNames());
+        return this.entity instanceof ServerPlayer sourcePlayer && !sourcePlayer.getBukkitEntity().hasPermission("paper.bypass-visibility.tab-completion") ? this.getServer().getPlayerList().getPlayers().stream().filter(serverPlayer -> sourcePlayer.getBukkitEntity().canSee(serverPlayer.getBukkitEntity())).map(serverPlayer -> serverPlayer.getGameProfile().name()).toList() : Lists.newArrayList(this.server.getPlayerNames()); // Paper - Make CommandSourceStack respect hidden players
     }
 
     @Override
@@ -528,21 +_,24 @@
     }
 
     @Override
-    public <E> CompletableFuture<Suggestions> suggestRegistryElements(
-        ResourceKey<? extends Registry<E>> key,
-        SharedSuggestionProvider.ElementSuggestionType elements,
-        SuggestionsBuilder builder,
-        CommandContext<?> context,
-        Predicate<E> filter
+    public CompletableFuture<Suggestions> suggestRegistryElements(
+        ResourceKey<? extends Registry<?>> key, SharedSuggestionProvider.ElementSuggestionType elements, SuggestionsBuilder builder, CommandContext<?> context
     ) {
-        return this.getLookup(key).map(registry -> {
-            this.suggestRegistryElements((HolderLookup<E>)registry, elements, builder, filter);
-            return builder.buildFuture();
-        }).orElseGet(Suggestions::empty);
+        if (key == Registries.RECIPE) {
+            return SharedSuggestionProvider.suggestResource(this.server.getRecipeManager().getRecipes().stream().map(e -> e.id().identifier()), builder);
+        } else if (key == Registries.ADVANCEMENT) {
+            Collection<AdvancementHolder> advancements = this.server.getAdvancements().getAllAdvancements();
+            return SharedSuggestionProvider.suggestResource(advancements.stream().map(AdvancementHolder::id), builder);
+        } else {
+            return this.getLookup(key).map(registry -> {
+                this.suggestRegistryElements((HolderLookup<?>)registry, elements, builder);
+                return builder.buildFuture();
+            }).orElseGet(Suggestions::empty);
+        }
     }
 
-    private <E> Optional<? extends HolderLookup<E>> getLookup(ResourceKey<? extends Registry<E>> key) {
-        Optional<? extends Registry<E>> lookup = this.registryAccess().lookup(key);
+    private Optional<? extends HolderLookup<?>> getLookup(ResourceKey<? extends Registry<?>> key) {
+        Optional<? extends Registry<?>> lookup = this.registryAccess().lookup(key);
         return lookup.isPresent() ? lookup : this.server.reloadableRegistries().lookup().lookup(key);
     }
 
@@ -582,35 +_,15 @@
         return this.silent;
     }
 
-    public interface NamesProvider {
-        CommandSourceStack.NamesProvider FOR_ENTITY = new CommandSourceStack.NamesProvider() {
-            @Override
-            public Component displayName(@Nullable Entity entity) {
-                return entity != null ? entity.getDisplayName() : CommonComponents.EMPTY;
-            }
-
-            @Override
-            public String textName(@Nullable Entity entity) {
-                return entity != null ? entity.getPlainTextName() : "";
-            }
-        };
-
-        Component displayName(@Nullable Entity entity);
-
-        String textName(@Nullable Entity entity);
-
-        static CommandSourceStack.NamesProvider constant(Component value) {
-            return new CommandSourceStack.NamesProvider() {
-                @Override
-                public Component displayName(@Nullable Entity entity) {
-                    return value;
-                }
-
-                @Override
-                public String textName(@Nullable Entity entity) {
-                    return value.getString();
-                }
-            };
-        }
-    }
+    // Paper start
+    @Override
+    public CommandSourceStack getHandle() {
+        return this;
+    }
+    // Paper end
+    // CraftBukkit start
+    public org.bukkit.command.CommandSender getBukkitSender() {
+        return this.source.getBukkitSender(this);
+    }
+    // CraftBukkit end
 }
