--- a/net/minecraft/commands/ExecutionCommandSource.java
+++ b/net/minecraft/commands/ExecutionCommandSource.java
@@ -3,7 +_,6 @@
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.Message;
 import com.mojang.brigadier.ResultConsumer;
-import com.mojang.brigadier.context.CommandContext;
 import com.mojang.brigadier.exceptions.CommandExceptionType;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import javax.annotation.Nullable;
