--- a/net/minecraft/client/model/geom/LayerDefinitions.java
+++ b/net/minecraft/client/model/geom/LayerDefinitions.java
@@ -162,8 +_,8 @@
 @OnlyIn(Dist.CLIENT)
 public class LayerDefinitions {
     private static final CubeDeformation FISH_PATTERN_DEFORMATION = new CubeDeformation(0.008F);
-    public static final CubeDeformation OUTER_ARMOR_DEFORMATION = new CubeDeformation(1.0F);
-    public static final CubeDeformation INNER_ARMOR_DEFORMATION = new CubeDeformation(0.5F);
+    private static final CubeDeformation OUTER_ARMOR_DEFORMATION = new CubeDeformation(1.0F);
+    private static final CubeDeformation INNER_ARMOR_DEFORMATION = new CubeDeformation(0.5F);
     private static final CubeDeformation BABY_OUTER_ARMOR_DEFORMATION = new CubeDeformation(-0.1F, 0.5F, 0.3F);
     private static final CubeDeformation BABY_INNER_ARMOR_DEFORMATION = new CubeDeformation(-0.1F, 0.3F, 0.3F);
     private static final CubeDeformation BABY_PIGLIN_INNER_ARMOR_DEFORMATION = new CubeDeformation(0.7F);
