--- a/net/minecraft/client/model/geom/ModelPart.java
+++ b/net/minecraft/client/model/geom/ModelPart.java
@@ -166,7 +_,7 @@
     public void translateAndRotate(PoseStack poseStack) {
         poseStack.translate(this.x / 16.0F, this.y / 16.0F, this.z / 16.0F);
         if (this.xRot != 0.0F || this.yRot != 0.0F || this.zRot != 0.0F) {
-            poseStack.rotate(new Quaternionf().rotationZYX(this.zRot, this.yRot, this.xRot));
+            poseStack.mulPose(new Quaternionf().rotationZYX(this.zRot, this.yRot, this.xRot));
         }
 
         if (this.xScale != 1.0F || this.yScale != 1.0F || this.zScale != 1.0F) {
