--- a/net/minecraft/world/waypoints/WaypointTransmitter.java
+++ b/net/minecraft/world/waypoints/WaypointTransmitter.java
@@ -20,20 +_,24 @@
     Waypoint.Icon waypointIcon();
 
     static boolean doesSourceIgnoreReceiver(LivingEntity source, ServerPlayer receiver) {
+        if (!receiver.getBukkitEntity().canSee(source.getBukkitEntity())) return true; // Paper - ignore if entity is hidden from player
         if (receiver.isSpectator()) {
             return false;
         } else if (!source.isSpectator() && !source.hasIndirectPassenger(receiver)) {
             double broadcastRange = Math.min(
                 source.getAttributeValue(Attributes.WAYPOINT_TRANSMIT_RANGE), receiver.getAttributeValue(Attributes.WAYPOINT_RECEIVE_RANGE)
             );
-            return source.distanceTo(receiver) >= broadcastRange;
+            return source.distanceToSqr(receiver) >= Mth.square(broadcastRange); // Paper - optimize waypoint distance check by using squared distance
         } else {
             return true;
         }
     }
 
     static boolean isChunkVisible(ChunkPos chunkPos, ServerPlayer receiver) {
-        return receiver.getChunkTrackingView().isInViewDistance(chunkPos.x(), chunkPos.z());
+        // Paper start - rewrite chunk system
+        final ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData playerChunkLoader = ((ca.spottedleaf.moonrise.patches.chunk_system.player.ChunkSystemServerPlayer)receiver).moonrise$getChunkLoader();
+        return playerChunkLoader != null && playerChunkLoader.getSentChunksRaw().contains(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkPos));
+        // Paper end - rewrite chunk system
     }
 
     static boolean isReallyFar(LivingEntity source, ServerPlayer receiver) {
