--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -266,7 +_,7 @@
 
     public abstract SoundEvent getCelebrateSound();
 
-    public class HoldGroundAttackGoal extends Goal {
+    protected class HoldGroundAttackGoal extends Goal {
         private final Raider mob;
         private final float hostileRadiusSqr;
         public final TargetingConditions shoutTargeting = TargetingConditions.forNonCombat().range(8.0).ignoreLineOfSight().ignoreInvisibilityTesting();
