--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -222,6 +_,7 @@
             this.onItemPickup(p_37866_);
             this.setItemSlot(equipmentslot, itemstack);
             this.take(p_37866_, itemstack.getCount());
+            p_37866_.removeReason(org.bukkit.event.entity.EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
             p_37866_.discard();
             this.getCurrentRaid().setLeader(this.getWave(), this);
             this.setPatrolLeader(true);
@@ -295,6 +_,7 @@
             for (Raider raider : this.mob
                 .level()
                 .getNearbyEntities(Raider.class, this.shoutTargeting, this.mob, this.mob.getBoundingBox().inflate(8.0, 8.0, 8.0))) {
+                raider.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                 raider.setTarget(this.mob.getTarget());
             }
         }
@@ -307,6 +_,7 @@
                 for (Raider raider : this.mob
                     .level()
                     .getNearbyEntities(Raider.class, this.shoutTargeting, this.mob, this.mob.getBoundingBox().inflate(8.0, 8.0, 8.0))) {
+                    raider.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                     raider.setTarget(livingentity);
                     raider.setAggressive(true);
                 }
@@ -348,6 +_,7 @@
 
         @Override
         public boolean canUse() {
+            if (!this.mob.level().getGameRules().getBoolean(net.minecraft.world.level.GameRules.RULE_MOBGRIEFING) || !this.mob.canPickUpLoot()) return false; // Paper - respect game and entity rules for picking up items
             Raid raid = this.mob.getCurrentRaid();
             if (this.mob.hasActiveRaid()
                 && !this.mob.getCurrentRaid().isOver()
