--- a/net/minecraft/world/entity/raid/Raids.java
+++ b/net/minecraft/world/entity/raid/Raids.java
@@ -132,12 +_,21 @@
         }
 
         Raid raid = this.getOrCreateRaid(level, raidCenterPos);
-        if (!raid.isStarted() && !this.raidMap.containsValue(raid)) {
-            this.raidMap.put(this.getUniqueId(), raid);
-        }
-
-        if (!raid.isStarted() || raid.getRaidOmenLevel() < raid.getMaxRaidOmenLevel()) {
-            raid.absorbRaidOmen(player);
+
+        if (!raid.isStarted() || (raid.isInProgress() && raid.getRaidOmenLevel() < raid.getMaxRaidOmenLevel())) { // CraftBukkit - fixed a bug with raid
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callRaidTriggerEvent(raid, level, player)) {
+                player.removeEffect(net.minecraft.world.effect.MobEffects.RAID_OMEN);
+                return null;
+            }
+
+            if (!raid.isStarted() && !this.raidMap.containsValue(raid)) {
+                this.raidMap.put(this.getUniqueId(), raid);
+            }
+            // CraftBukkit end
+            if (!raid.isStarted() || raid.getRaidOmenLevel() < raid.getMaxRaidOmenLevel()) {
+                raid.absorbRaidOmen(player);
+            }
         }
 
         this.setDirty();
