--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -185,6 +_,12 @@
         return this.status == Raid.RaidStatus.LOSS;
     }
 
+    // CraftBukkit start
+    public boolean isInProgress() {
+        return this.status == RaidStatus.ONGOING;
+    }
+    // CraftBukkit end
+
     public float getTotalHealth() {
         return this.totalHealth;
     }
@@ -271,6 +_,7 @@
                 boolean oldActive = this.active;
                 this.active = level.hasChunkAt(this.center);
                 if (level.getDifficulty() == Difficulty.PEACEFUL) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.PEACE); // CraftBukkit
                     this.stop();
                     return;
                 }
@@ -290,13 +_,16 @@
                 if (!level.isVillage(this.center)) {
                     if (this.groupsSpawned > 0) {
                         this.status = Raid.RaidStatus.LOSS;
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidFinishEvent(this, level, new java.util.ArrayList<>()); // CraftBukkit
                     } else {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.NOT_IN_VILLAGE); // CraftBukkit
                         this.stop();
                     }
                 }
 
                 this.ticksActive++;
                 if (this.ticksActive >= 48000L) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.TIMEOUT); // CraftBukkit
                     this.stop();
                     return;
                 }
@@ -385,6 +_,7 @@
                     }
 
                     if (attempt > 5) {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.UNSPAWNABLE);  // CraftBukkit
                         this.stop();
                         break;
                     }
@@ -396,6 +_,7 @@
                     } else {
                         this.status = Raid.RaidStatus.VICTORY;
 
+                        List<org.bukkit.entity.Player> winners = new java.util.ArrayList<>(); // CraftBukkit
                         for (UUID heroUUID : this.heroesOfTheVillage) {
                             Entity entity = level.getEntity(heroUUID);
                             if (entity instanceof LivingEntity hero && !entity.isSpectator()) {
@@ -403,9 +_,11 @@
                                 if (hero instanceof ServerPlayer playerHero) {
                                     playerHero.awardStat(Stats.RAID_WIN);
                                     CriteriaTriggers.RAID_WIN.trigger(playerHero);
+                                    winners.add(playerHero.getBukkitEntity()); // CraftBukkit
                                 }
                             }
                         }
+                        org.bukkit.craftbukkit.event.CraftEventFactory.callRaidFinishEvent(this, level, winners); // CraftBukkit
                     }
                 }
 
@@ -413,6 +_,7 @@
             } else if (this.isOver()) {
                 this.celebrationTicks++;
                 if (this.celebrationTicks >= 600) {
+                    org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(this, level, org.bukkit.event.raid.RaidStopEvent.Reason.FINISHED); // CraftBukkit
                     this.stop();
                     return;
                 }
@@ -530,7 +_,7 @@
             int ravagersSpawned = 0;
 
             for (int i = 0; i < numSpawns; i++) {
-                Raider raider = raiderType.entityType.create(level, EntitySpawnReason.EVENT);
+                Raider raider = raiderType.entityTypeSupplier.get().create(level, EntitySpawnReason.EVENT);
                 if (raider == null) {
                     break;
                 }
@@ -542,7 +_,7 @@
                 }
 
                 this.joinRaid(level, groupNumber, raider, pos, false);
-                if (raiderType.entityType == EntityType.RAVAGER) {
+                if (raiderType.entityTypeSupplier.get() == EntityType.RAVAGER) {
                     Raider ridingRaider = null;
                     if (groupNumber == this.getNumGroups(Difficulty.NORMAL)) {
                         ridingRaider = EntityType.PILLAGER.create(level, EntitySpawnReason.EVENT);
@@ -582,7 +_,7 @@
                 raider.finalizeSpawn(level, level.getCurrentDifficultyAt(pos), EntitySpawnReason.EVENT, null);
                 raider.applyRaidBuffs(level, groupNumber, false);
                 raider.setOnGround(true);
-                level.addFreshEntityWithPassengers(raider);
+                level.addFreshEntityWithPassengers(raider, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.RAID); // CraftBukkit
             }
         }
     }
@@ -808,6 +_,12 @@
         this.heroesOfTheVillage.add(killer.getUUID());
     }
 
+    // CraftBukkit start - a method to get all raiders
+    public java.util.Collection<Raider> getRaiders() {
+        return this.groupRaiderMap.values().stream().flatMap(Set::stream).collect(java.util.stream.Collectors.toSet());
+    }
+    // CraftBukkit end
+
     private static enum RaidStatus implements StringRepresentable {
         ONGOING("ongoing"),
         VICTORY("victory"),
@@ -835,12 +_,26 @@
         RAVAGER(EntityType.RAVAGER, new int[]{0, 0, 0, 1, 0, 1, 0, 2});
 
         private static final Raid.RaiderType[] VALUES = values();
+        @Deprecated // Neo: null for custom types, use the supplier instead
         private final EntityType<? extends Raider> entityType;
         private final int[] spawnsPerWaveBeforeBonus;
+        final java.util.function.Supplier<EntityType<? extends Raider>> entityTypeSupplier;
 
+        @net.neoforged.fml.common.asm.enumextension.ReservedConstructor
         private RaiderType(EntityType<? extends Raider> entityType, int[] spawnsPerWaveBeforeBonus) {
             this.entityType = entityType;
             this.spawnsPerWaveBeforeBonus = spawnsPerWaveBeforeBonus;
+            this.entityTypeSupplier = () -> entityType;
+        }
+
+        private RaiderType(java.util.function.Supplier<EntityType<? extends Raider>> entityTypeSupplier, int[] spawnsPerWave) {
+            this.entityType = null;
+            this.spawnsPerWaveBeforeBonus = spawnsPerWave;
+            this.entityTypeSupplier = entityTypeSupplier;
+        }
+
+        public static net.neoforged.fml.common.asm.enumextension.ExtensionInfo getExtensionInfo() {
+            return net.neoforged.fml.common.asm.enumextension.ExtensionInfo.nonExtended(Raid.RaiderType.class);
         }
     }
 }
