--- a/net/minecraft/world/level/levelgen/BelowZeroRetrogen.java
+++ b/net/minecraft/world/level/levelgen/BelowZeroRetrogen.java
@@ -101,8 +_,8 @@
         }
 
         Predicate<ResourceKey<Biome>> retainedBiomes = RETAINED_RETROGEN_BIOMES::contains;
-        return (quartX, quartY, quartZ) -> {
-            Holder<Biome> noiseBiome = biomeResolver.getNoiseBiome(quartX, quartY, quartZ);
+        return (quartX, quartY, quartZ, sampler) -> {
+            Holder<Biome> noiseBiome = biomeResolver.getNoiseBiome(quartX, quartY, quartZ, sampler);
             return noiseBiome.is(retainedBiomes) ? noiseBiome : protoChunk.getNoiseBiome(quartX, 0, quartZ);
         };
     }
