--- a/net/minecraft/world/level/levelgen/BitRandomSource.java
+++ b/net/minecraft/world/level/levelgen/BitRandomSource.java
@@ -16,7 +_,7 @@
     @Override
     default int nextInt(int p_188504_) {
         if (p_188504_ <= 0) {
-            throw new IllegalArgumentException("Bound must be positive");
+            throw new IllegalArgumentException("Bound must be positive, got: " + p_188504_);
         } else if ((p_188504_ & p_188504_ - 1) == 0) {
             return (int)((long)p_188504_ * (long)this.next(31) >> 31);
         } else {
