--- a/net/minecraft/world/level/levelgen/DensityFunctions.java
+++ b/net/minecraft/world/level/levelgen/DensityFunctions.java
@@ -20,7 +_,6 @@
 import net.minecraft.util.BoundedFloatFunction;
 import net.minecraft.util.CubicSpline;
 import net.minecraft.util.ExtraCodecs;
-import net.minecraft.util.Interval;
 import net.minecraft.util.KeyDispatchDataCodec;
 import net.minecraft.util.Mth;
 import net.minecraft.util.RandomSource;
@@ -228,7 +_,7 @@
     }
 
     public static DensityFunction map(DensityFunction function, DensityFunctions.Mapped.Type type) {
-        return new DensityFunctions.Mapped(type, function);
+        return DensityFunctions.Mapped.create(type, function);
     }
 
     private static DensityFunction mapFromUnitTo(DensityFunction function, double min, double max) {
@@ -264,17 +_,8 @@
     }
 
     private record Ap2(
-        DensityFunctions.TwoArgumentSimpleFunction.Type type, DensityFunction argument1, DensityFunction argument2, double minValue2, double maxValue2
+        DensityFunctions.TwoArgumentSimpleFunction.Type type, DensityFunction argument1, DensityFunction argument2, double minValue, double maxValue
     ) implements DensityFunctions.TwoArgumentSimpleFunction {
-        public Ap2(DensityFunctions.TwoArgumentSimpleFunction.Type type, DensityFunction argument1, DensityFunction argument2) {
-            Interval range2 = argument2.range();
-            this(type, argument1, argument2, range2.min(), range2.max());
-            if ((type == DensityFunctions.TwoArgumentSimpleFunction.Type.MIN || type == DensityFunctions.TwoArgumentSimpleFunction.Type.MAX)
-                && !argument1.range().intersects(range2)) {
-                LOGGER.warn("Creating a {} function between two non-overlapping inputs: {} and {}", type, argument1, argument2);
-            }
-        }
-
         @Override
         public double compute(DensityFunction.FunctionContext context) {
             double v1 = this.argument1.compute(context);
@@ -282,8 +_,8 @@
             return switch (this.type) {
                 case ADD -> v1 + this.argument2.compute(context);
                 case MUL -> v1 == 0.0 ? 0.0 : v1 * this.argument2.compute(context);
-                case MIN -> v1 < this.minValue2 ? v1 : Math.min(v1, this.argument2.compute(context));
-                case MAX -> v1 > this.maxValue2 ? v1 : Math.max(v1, this.argument2.compute(context));
+                case MIN -> v1 < this.argument2.minValue() ? v1 : Math.min(v1, this.argument2.compute(context));
+                case MAX -> v1 > this.argument2.maxValue() ? v1 : Math.max(v1, this.argument2.compute(context));
             };
         }
 
@@ -306,15 +_,19 @@
                     }
                     break;
                 case MIN:
+                    double min = this.argument2.minValue();
+
                     for (int i = 0; i < output.length; i++) {
                         double v = output[i];
-                        output[i] = v < this.minValue2 ? v : Math.min(v, this.argument2.compute(contextProvider.forIndex(i)));
+                        output[i] = v < min ? v : Math.min(v, this.argument2.compute(contextProvider.forIndex(i)));
                     }
                     break;
                 case MAX:
+                    double max = this.argument2.maxValue();
+
                     for (int i = 0; i < output.length; i++) {
                         double v = output[i];
-                        output[i] = v > this.maxValue2 ? v : Math.max(v, this.argument2.compute(contextProvider.forIndex(i)));
+                        output[i] = v > max ? v : Math.max(v, this.argument2.compute(contextProvider.forIndex(i)));
                     }
             }
         }
@@ -323,19 +_,6 @@
         public DensityFunction mapChildren(DensityFunction.Visitor visitor) {
             return DensityFunctions.TwoArgumentSimpleFunction.create(this.type, visitor.apply(this.argument1), visitor.apply(this.argument2));
         }
-
-        @Override
-        public Interval range() {
-            Interval range1 = this.argument1.range();
-            Interval range2 = this.argument2.range();
-
-            return switch (this.type) {
-                case ADD -> Interval.add(range1, range2);
-                case MUL -> Interval.mul(range1, range2);
-                case MIN -> Interval.min(range1, range2);
-                case MAX -> Interval.max(range1, range2);
-            };
-        }
     }
 
     enum BeardifierMarker implements DensityFunctions.BeardifierOrMarker {
@@ -350,17 +_,22 @@
         public void fillArray(double[] output, DensityFunction.ContextProvider contextProvider) {
             Arrays.fill(output, 0.0);
         }
+
+        @Override
+        public double minValue() {
+            return 0.0;
+        }
+
+        @Override
+        public double maxValue() {
+            return 0.0;
+        }
     }
 
     public interface BeardifierOrMarker extends DensityFunction.SimpleFunction {
         KeyDispatchDataCodec<DensityFunction> CODEC = KeyDispatchDataCodec.of(MapCodec.unit(DensityFunctions.BeardifierMarker.INSTANCE));
 
         @Override
-        default Interval range() {
-            return Beardifier.RANGE;
-        }
-
-        @Override
         default KeyDispatchDataCodec<? extends DensityFunction> codec() {
             return CODEC;
         }
@@ -382,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.of(0.0, 1.0);
+        public double minValue() {
+            return 1.0;
+        }
+
+        @Override
+        public double maxValue() {
+            return 1.0;
         }
 
         @Override
@@ -408,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.INFINITE;
+        public double minValue() {
+            return 0.0;
+        }
+
+        @Override
+        public double maxValue() {
+            return 0.0;
         }
 
         @Override
@@ -418,42 +_,30 @@
         }
     }
 
-    protected record Clamp(DensityFunction input, double min, double max) implements DensityFunctions.PureTransformer {
-        private static final MapCodec<DensityFunctions.Clamp> DATA_CODEC = RecordCodecBuilder.<DensityFunctions.Clamp>mapCodec(
-                i -> i.group(
-                        DensityFunction.CODEC.fieldOf("input").forGetter(DensityFunctions.Clamp::input),
-                        DensityFunctions.NOISE_VALUE_CODEC.fieldOf("min").forGetter(DensityFunctions.Clamp::min),
-                        DensityFunctions.NOISE_VALUE_CODEC.fieldOf("max").forGetter(DensityFunctions.Clamp::max)
-                    )
-                    .apply(i, DensityFunctions.Clamp::new)
-            )
-            .validate(DensityFunctions.Clamp::validate);
+    protected record Clamp(DensityFunction input, double minValue, double maxValue) implements DensityFunctions.PureTransformer {
+        private static final MapCodec<DensityFunctions.Clamp> DATA_CODEC = RecordCodecBuilder.mapCodec(
+            i -> i.group(
+                    DensityFunction.CODEC.fieldOf("input").forGetter(DensityFunctions.Clamp::input),
+                    DensityFunctions.NOISE_VALUE_CODEC.fieldOf("min").forGetter(DensityFunctions.Clamp::minValue),
+                    DensityFunctions.NOISE_VALUE_CODEC.fieldOf("max").forGetter(DensityFunctions.Clamp::maxValue)
+                )
+                .apply(i, DensityFunctions.Clamp::new)
+        );
         public static final KeyDispatchDataCodec<DensityFunctions.Clamp> CODEC = DensityFunctions.makeCodec(DATA_CODEC);
 
-        private static DataResult<DensityFunctions.Clamp> validate(DensityFunctions.Clamp clamp) {
-            return clamp.max < clamp.min
-                ? DataResult.error(() -> "min (" + clamp.min + ") must be less than or equal to max (" + clamp.max + ")")
-                : DataResult.success(clamp);
-        }
-
         @Override
         public double transform(double input) {
-            return Mth.clamp(input, this.min, this.max);
+            return Mth.clamp(input, this.minValue, this.maxValue);
         }
 
         @Override
         public DensityFunction mapChildren(DensityFunction.Visitor visitor) {
-            return new DensityFunctions.Clamp(visitor.apply(this.input), this.min, this.max);
+            return new DensityFunctions.Clamp(visitor.apply(this.input), this.minValue, this.maxValue);
         }
 
         @Override
         public KeyDispatchDataCodec<? extends DensityFunction> codec() {
             return CODEC;
-        }
-
-        @Override
-        public Interval range() {
-            return Interval.clamp(this.input.range(), this.min, this.max);
         }
     }
 
@@ -474,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.ofExact(this.value);
+        public double minValue() {
+            return this.value;
+        }
+
+        @Override
+        public double maxValue() {
+            return this.value;
         }
 
         @Override
@@ -490,6 +_,16 @@
         );
         private static final float ISLAND_THRESHOLD = -0.9F;
         private final SimplexNoise islandNoise;
+        // Paper start - Perf: Optimize end generation
+        private static final class NoiseCache {
+            public long[] keys = new long[8192];
+            public float[] values = new float[8192];
+            public NoiseCache() {
+                java.util.Arrays.fill(keys, Long.MIN_VALUE);
+            }
+        }
+        private static final ThreadLocal<java.util.Map<SimplexNoise, NoiseCache>> noiseCache = ThreadLocal.withInitial(java.util.WeakHashMap::new);
+        // Paper end - Perf: Optimize end generation
 
         public EndIslandDensityFunction(long seed) {
             RandomSource islandRandom = new LegacyRandomSource(seed);
@@ -502,15 +_,31 @@
             int chunkZ = sectionZ / 2;
             int subSectionX = sectionX % 2;
             int subSectionZ = sectionZ % 2;
-            float doffs = 100.0F - Mth.sqrt(sectionX * sectionX + sectionZ * sectionZ) * 8.0F;
+            float doffs = 100.0F - (ca.spottedleaf.moonrise.common.PlatformHooks.get().configFixMC159283() ? Mth.sqrt((long) sectionX * (long) sectionX + (long) sectionZ * (long) sectionZ) : Mth.sqrt(sectionX * sectionX + sectionZ * sectionZ)) * 8.0F; // Paper - cast ints to long when MC-159283 fix enabled
             doffs = Mth.clamp(doffs, -100.0F, 80.0F);
 
+            NoiseCache cache = noiseCache.get().computeIfAbsent(islandNoise, noiseKey -> new NoiseCache()); // Paper - Perf: Optimize end generation
             for (int xo = -12; xo <= 12; xo++) {
                 for (int zo = -12; zo <= 12; zo++) {
                     long totalChunkX = chunkX + xo;
                     long totalChunkZ = chunkZ + zo;
+                    // Paper start - Perf: Optimize end generation by using a noise cache
+                    final long chunkKey = net.minecraft.world.level.ChunkPos.pack((int) totalChunkX, (int) totalChunkZ);
+                    final int cacheIndex = (int) it.unimi.dsi.fastutil.HashCommon.mix(chunkKey) & 8191;
+                    float islandSize = Float.MIN_VALUE; // noise value
+                    if (cache.keys[cacheIndex] == chunkKey) {
+                        // Use cache
+                        islandSize = cache.values[cacheIndex];
+                    } else {
+                        // Vanilla function
                     if (totalChunkX * totalChunkX + totalChunkZ * totalChunkZ > 4096L && islandNoise.getValue(totalChunkX, totalChunkZ) < -0.9F) {
-                        float islandSize = (Mth.abs((float)totalChunkX) * 3439.0F + Mth.abs((float)totalChunkZ) * 147.0F) % 13.0F + 9.0F;
+                        islandSize = (Mth.abs((float)totalChunkX) * 3439.0F + Mth.abs((float) totalChunkZ) * 147.0F) % 13.0F + 9.0F;
+                    }
+                        cache.keys[cacheIndex] = chunkKey;
+                        cache.values[cacheIndex] = islandSize;
+                    }
+                    if (islandSize != Float.MIN_VALUE) {
+                        // Paper end - Perf: Optimize end generation
                         float xd = subSectionX - xo * 2;
                         float zd = subSectionZ - zo * 2;
                         float newDoffs = 100.0F - Mth.sqrt(xd * xd + zd * zd) * islandSize;
@@ -529,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.of(-0.84375, 0.5625);
+        public double minValue() {
+            return -0.84375;
+        }
+
+        @Override
+        public double maxValue() {
+            return 0.5625;
         }
 
         @Override
@@ -580,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.of(this.lowerBound, Math.max(this.lowerBound, this.upperBound.range().max()));
+        public double minValue() {
+            return this.lowerBound;
+        }
+
+        @Override
+        public double maxValue() {
+            return Math.max(this.lowerBound, this.upperBound.maxValue());
         }
 
         @Override
@@ -608,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return this.function.isBound() ? this.function.value().range() : Interval.INFINITE;
+        public double minValue() {
+            return this.function.isBound() ? this.function.value().minValue() : Double.NEGATIVE_INFINITY;
+        }
+
+        @Override
+        public double maxValue() {
+            return this.function.isBound() ? this.function.value().maxValue() : Double.POSITIVE_INFINITY;
         }
 
         @Override
@@ -678,8 +_,25 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.encapsulating(Lists.transform(this.functions, DensityFunction::range));
+        public double minValue() {
+            double minValue = Double.MAX_VALUE;
+
+            for (DensityFunction function : this.functions) {
+                minValue = Math.min(function.minValue(), minValue);
+            }
+
+            return minValue;
+        }
+
+        @Override
+        public double maxValue() {
+            double maxValue = -Double.MAX_VALUE;
+
+            for (DensityFunction function : this.functions) {
+                maxValue = Math.max(function.maxValue(), maxValue);
+            }
+
+            return maxValue;
         }
 
         @Override
@@ -688,7 +_,24 @@
         }
     }
 
-    protected record Mapped(DensityFunctions.Mapped.Type type, DensityFunction input) implements DensityFunctions.PureTransformer {
+    protected record Mapped(DensityFunctions.Mapped.Type type, DensityFunction input, double minValue, double maxValue)
+        implements DensityFunctions.PureTransformer {
+        public static DensityFunctions.Mapped create(DensityFunctions.Mapped.Type type, DensityFunction input) {
+            double minValue = input.minValue();
+            double maxValue = input.maxValue();
+            double minImage = transform(type, minValue);
+            double maxImage = transform(type, maxValue);
+            if (type == DensityFunctions.Mapped.Type.INVERT) {
+                return minValue < 0.0 && maxValue > 0.0
+                    ? new DensityFunctions.Mapped(type, input, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY)
+                    : new DensityFunctions.Mapped(type, input, maxImage, minImage);
+            } else {
+                return type != DensityFunctions.Mapped.Type.ABS && type != DensityFunctions.Mapped.Type.SQUARE
+                    ? new DensityFunctions.Mapped(type, input, minImage, maxImage)
+                    : new DensityFunctions.Mapped(type, input, Math.max(0.0, minValue), Math.max(minImage, maxImage));
+            }
+        }
+
         private static double transform(DensityFunctions.Mapped.Type type, double input) {
             return switch (type) {
                 case ABS -> Math.abs(input);
@@ -710,7 +_,7 @@
         }
 
         public DensityFunctions.Mapped mapChildren(DensityFunction.Visitor visitor) {
-            return new DensityFunctions.Mapped(this.type, visitor.apply(this.input));
+            return create(this.type, visitor.apply(this.input));
         }
 
         @Override
@@ -718,18 +_,6 @@
             return this.type.codec;
         }
 
-        @Override
-        public Interval range() {
-            Interval input = this.input.range();
-
-            return switch (this.type) {
-                case ABS -> Interval.abs(input);
-                case SQUARE -> Interval.square(input);
-                case CUBE, HALF_NEGATIVE, QUARTER_NEGATIVE, SQUEEZE -> Interval.mapMonotonic(input, value -> transform(this.type, value));
-                case INVERT -> Interval.inverse(input);
-            };
-        }
-
         public enum Type implements StringRepresentable {
             ABS("abs"),
             SQUARE("square"),
@@ -741,7 +_,7 @@
 
             private final String name;
             private final KeyDispatchDataCodec<DensityFunctions.Mapped> codec = DensityFunctions.singleFunctionArgumentCodec(
-                input -> new DensityFunctions.Mapped(this, input), DensityFunctions.Mapped::input
+                input -> DensityFunctions.Mapped.create(this, input), DensityFunctions.Mapped::input
             );
 
             Type(String name) {
@@ -767,11 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return switch (this.type) {
-                case Interpolated, FlatCache, Cache2D, CacheOnce, CacheAllInCell -> this.wrapped.range();
-                case BlendDensity -> Interval.INFINITE;
-            };
+        public double minValue() {
+            return this.type == DensityFunctions.Marker.Type.BlendDensity ? Double.NEGATIVE_INFINITY : this.wrapped.minValue();
+        }
+
+        @Override
+        public double maxValue() {
+            return this.type == DensityFunctions.Marker.Type.BlendDensity ? Double.POSITIVE_INFINITY : this.wrapped.maxValue();
         }
 
         public enum Type implements StringRepresentable {
@@ -814,7 +_,7 @@
         }
     }
 
-    private record MulOrAdd(DensityFunctions.MulOrAdd.Type specificType, DensityFunction input, double argument)
+    private record MulOrAdd(DensityFunctions.MulOrAdd.Type specificType, DensityFunction input, double minValue, double maxValue, double argument)
         implements DensityFunctions.TwoArgumentSimpleFunction,
         DensityFunctions.PureTransformer {
         @Override
@@ -844,15 +_,23 @@
 
         @Override
         public DensityFunction mapChildren(DensityFunction.Visitor visitor) {
-            return new DensityFunctions.MulOrAdd(this.specificType, visitor.apply(this.input), this.argument);
-        }
+            DensityFunction function = visitor.apply(this.input);
+            double min = function.minValue();
+            double max = function.maxValue();
+            double minValue;
+            double maxValue;
+            if (this.specificType == DensityFunctions.MulOrAdd.Type.ADD) {
+                minValue = min + this.argument;
+                maxValue = max + this.argument;
+            } else if (this.argument >= 0.0) {
+                minValue = min * this.argument;
+                maxValue = max * this.argument;
+            } else {
+                minValue = max * this.argument;
+                maxValue = min * this.argument;
+            }
 
-        @Override
-        public Interval range() {
-            return switch (this.specificType) {
-                case MUL -> Interval.mul(this.input.range(), Interval.ofExact(this.argument));
-                case ADD -> Interval.add(this.input.range(), Interval.ofExact(this.argument));
-            };
+            return new DensityFunctions.MulOrAdd(this.specificType, function, minValue, maxValue, this.argument);
         }
 
         public enum Type {
@@ -888,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.ofSymmetric(this.noise.maxValue());
+        public double minValue() {
+            return -this.maxValue();
+        }
+
+        @Override
+        public double maxValue() {
+            return this.noise.maxValue();
         }
 
         @Override
@@ -960,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.encapsulating(this.whenInRange.range(), this.whenOutOfRange.range());
+        public double minValue() {
+            return Math.min(this.whenInRange.minValue(), this.whenOutOfRange.minValue());
+        }
+
+        @Override
+        public double maxValue() {
+            return Math.max(this.whenInRange.maxValue(), this.whenOutOfRange.maxValue());
         }
 
         @Override
@@ -1037,8 +_,13 @@
         DensityFunction.NoiseHolder offsetNoise();
 
         @Override
-        default Interval range() {
-            return Interval.ofSymmetric(this.offsetNoise().maxValue() * 4.0);
+        default double minValue() {
+            return -this.maxValue();
+        }
+
+        @Override
+        default double maxValue() {
+            return this.offsetNoise().maxValue() * 4.0;
         }
 
         default double compute(double localX, double localY, double localZ) {
@@ -1088,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.ofSymmetric(this.noise.maxValue());
+        public double minValue() {
+            return -this.maxValue();
+        }
+
+        @Override
+        public double maxValue() {
+            return this.noise.maxValue();
         }
 
         @Override
@@ -1117,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return this.spline.range();
+        public double minValue() {
+            return this.spline.minValue();
+        }
+
+        @Override
+        public double maxValue() {
+            return this.spline.maxValue();
         }
 
         @Override
@@ -1164,8 +_,13 @@
             }
 
             @Override
-            public Interval range() {
-                return this.function.range();
+            public float minValue() {
+                return (float)this.function.minValue();
+            }
+
+            @Override
+            public float maxValue() {
+                return (float)this.function.maxValue();
             }
 
             public DensityFunctions.Spline.Coordinate mapChildren(DensityFunction.Visitor visitor) {
@@ -1203,11 +_,37 @@
         static DensityFunctions.TwoArgumentSimpleFunction create(
             DensityFunctions.TwoArgumentSimpleFunction.Type type, DensityFunction argument1, DensityFunction argument2
         ) {
+            double min1 = argument1.minValue();
+            double min2 = argument2.minValue();
+            double max1 = argument1.maxValue();
+            double max2 = argument2.maxValue();
+            if (type == DensityFunctions.TwoArgumentSimpleFunction.Type.MIN || type == DensityFunctions.TwoArgumentSimpleFunction.Type.MAX) {
+                boolean firstAlwaysBiggerThanSecond = min1 >= max2;
+                boolean secondAlwaysBiggerThanFirst = min2 >= max1;
+                if (firstAlwaysBiggerThanSecond || secondAlwaysBiggerThanFirst) {
+                    LOGGER.warn("Creating a {} function between two non-overlapping inputs: {} and {}", type, argument1, argument2);
+                }
+            }
+            double minValue = switch (type) {
+                case ADD -> min1 + min2;
+                case MUL -> min1 > 0.0 && min2 > 0.0 ? min1 * min2 : (max1 < 0.0 && max2 < 0.0 ? max1 * max2 : Math.min(min1 * max2, max1 * min2));
+                case MIN -> Math.min(min1, min2);
+                case MAX -> Math.max(min1, min2);
+            };
+
+            double maxValue = switch (type) {
+                case ADD -> max1 + max2;
+                case MUL -> min1 > 0.0 && min2 > 0.0 ? max1 * max2 : (max1 < 0.0 && max2 < 0.0 ? min1 * min2 : Math.max(min1 * min2, max1 * max2));
+                case MIN -> Math.min(max1, max2);
+                case MAX -> Math.max(max1, max2);
+            };
             if (type == DensityFunctions.TwoArgumentSimpleFunction.Type.MUL || type == DensityFunctions.TwoArgumentSimpleFunction.Type.ADD) {
                 if (argument1 instanceof DensityFunctions.Constant constant) {
                     return new DensityFunctions.MulOrAdd(
                         type == DensityFunctions.TwoArgumentSimpleFunction.Type.ADD ? DensityFunctions.MulOrAdd.Type.ADD : DensityFunctions.MulOrAdd.Type.MUL,
                         argument2,
+                        minValue,
+                        maxValue,
                         constant.value
                     );
                 }
@@ -1216,12 +_,14 @@
                     return new DensityFunctions.MulOrAdd(
                         type == DensityFunctions.TwoArgumentSimpleFunction.Type.ADD ? DensityFunctions.MulOrAdd.Type.ADD : DensityFunctions.MulOrAdd.Type.MUL,
                         argument1,
+                        minValue,
+                        maxValue,
                         constant.value
                     );
                 }
             }
 
-            return new DensityFunctions.Ap2(type, argument1, argument2);
+            return new DensityFunctions.Ap2(type, argument1, argument2, minValue, maxValue);
         }
 
         DensityFunctions.TwoArgumentSimpleFunction.Type type();
@@ -1277,8 +_,13 @@
         }
 
         @Override
-        public Interval range() {
-            return Interval.encapsulating(this.fromValue, this.toValue);
+        public double minValue() {
+            return Math.min(this.fromValue, this.toValue);
+        }
+
+        @Override
+        public double maxValue() {
+            return Math.max(this.fromValue, this.toValue);
         }
 
         @Override
