--- a/net/minecraft/world/level/levelgen/DensityFunctions.java
+++ b/net/minecraft/world/level/levelgen/DensityFunctions.java
@@ -521,7 +_,7 @@
             int j = p_224065_ / 2;
             int k = p_224064_ % 2;
             int l = p_224065_ % 2;
-            float f = 100.0F - Mth.sqrt((float)(p_224064_ * p_224064_ + p_224065_ * p_224065_)) * 8.0F;
+            float f = 100.0F - Mth.sqrt((long) p_224064_ * (long) p_224064_ + (long) p_224065_ * (long) p_224065_) * 8.0F; // Paper - cast ints to long to avoid integer overflow
             f = Mth.clamp(f, -100.0F, 80.0F);
 
             for (int i1 = -12; i1 <= 12; i1++) {
