--- a/net/minecraft/world/level/levelgen/Density.java
+++ b/net/minecraft/world/level/levelgen/Density.java
@@ -1,7 +_,7 @@
 package net.minecraft.world.level.levelgen;
 
 public class Density {
-    public static final float SURFACE = 0.0F;
-    public static final float UNRECOVERABLY_DENSE = 64.0F;
-    public static final float UNRECOVERABLY_THIN = -64.0F;
+    public static final double SURFACE = 0.0;
+    public static final double UNRECOVERABLY_DENSE = 64.0;
+    public static final double UNRECOVERABLY_THIN = -64.0;
 }
