--- a/net/minecraft/world/level/levelgen/GeodeBlockSettings.java
+++ b/net/minecraft/world/level/levelgen/GeodeBlockSettings.java
@@ -3,21 +_,20 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import java.util.List;
-import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.levelgen.feature.stateproviders.BlockStateProvider;
 
 public record GeodeBlockSettings(
-    Holder<BlockStateProvider> fillingProvider,
-    Holder<BlockStateProvider> innerLayerProvider,
-    Holder<BlockStateProvider> alternateInnerLayerProvider,
-    Holder<BlockStateProvider> middleLayerProvider,
-    Holder<BlockStateProvider> outerLayerProvider,
+    BlockStateProvider fillingProvider,
+    BlockStateProvider innerLayerProvider,
+    BlockStateProvider alternateInnerLayerProvider,
+    BlockStateProvider middleLayerProvider,
+    BlockStateProvider outerLayerProvider,
     List<BlockState> innerPlacements,
     HolderSet<Block> cannotReplace,
     HolderSet<Block> invalidBlocks
@@ -30,8 +_,8 @@
                 BlockStateProvider.CODEC.fieldOf("middle_layer_provider").forGetter(GeodeBlockSettings::middleLayerProvider),
                 BlockStateProvider.CODEC.fieldOf("outer_layer_provider").forGetter(GeodeBlockSettings::outerLayerProvider),
                 ExtraCodecs.nonEmptyList(BlockState.CODEC.listOf()).fieldOf("inner_placements").forGetter(GeodeBlockSettings::innerPlacements),
-                RegistryCodecs.holderSet(Registries.BLOCK).fieldOf("cannot_replace").forGetter(GeodeBlockSettings::cannotReplace),
-                RegistryCodecs.holderSet(Registries.BLOCK).fieldOf("invalid_blocks").forGetter(GeodeBlockSettings::invalidBlocks)
+                RegistryCodecs.homogeneousList(Registries.BLOCK).fieldOf("cannot_replace").forGetter(GeodeBlockSettings::cannotReplace),
+                RegistryCodecs.homogeneousList(Registries.BLOCK).fieldOf("invalid_blocks").forGetter(GeodeBlockSettings::invalidBlocks)
             )
             .apply(i, GeodeBlockSettings::new)
     );
