--- a/net/minecraft/world/level/levelgen/NoiseSettings.java
+++ b/net/minecraft/world/level/levelgen/NoiseSettings.java
@@ -4,23 +_,26 @@
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import java.util.function.Function;
+import net.minecraft.core.QuartPos;
 import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.dimension.DimensionType;
 
-public record NoiseSettings(int minY, int height) {
+public record NoiseSettings(int minY, int height, int noiseSizeHorizontal, int noiseSizeVertical) {
     public static final Codec<NoiseSettings> CODEC = RecordCodecBuilder.<NoiseSettings>create(
             i -> i.group(
                     Codec.intRange(DimensionType.MIN_Y, DimensionType.MAX_Y).fieldOf("min_y").forGetter(NoiseSettings::minY),
-                    Codec.intRange(0, DimensionType.Y_SIZE).fieldOf("height").forGetter(NoiseSettings::height)
+                    Codec.intRange(0, DimensionType.Y_SIZE).fieldOf("height").forGetter(NoiseSettings::height),
+                    Codec.intRange(1, 4).fieldOf("size_horizontal").forGetter(NoiseSettings::noiseSizeHorizontal),
+                    Codec.intRange(1, 4).fieldOf("size_vertical").forGetter(NoiseSettings::noiseSizeVertical)
                 )
                 .apply(i, NoiseSettings::new)
         )
         .comapFlatMap(NoiseSettings::guardY, Function.identity());
-    static final NoiseSettings OVERWORLD_NOISE_SETTINGS = create(-64, 384);
-    static final NoiseSettings NETHER_NOISE_SETTINGS = create(0, 128);
-    static final NoiseSettings END_NOISE_SETTINGS = create(0, 128);
-    static final NoiseSettings CAVES_NOISE_SETTINGS = create(-64, 192);
-    static final NoiseSettings FLOATING_ISLANDS_NOISE_SETTINGS = create(0, 256);
+    static final NoiseSettings OVERWORLD_NOISE_SETTINGS = create(-64, 384, 1, 2);
+    static final NoiseSettings NETHER_NOISE_SETTINGS = create(0, 128, 1, 2);
+    static final NoiseSettings END_NOISE_SETTINGS = create(0, 128, 2, 1);
+    static final NoiseSettings CAVES_NOISE_SETTINGS = create(-64, 192, 1, 2);
+    static final NoiseSettings FLOATING_ISLANDS_NOISE_SETTINGS = create(0, 256, 2, 1);
 
     private static DataResult<NoiseSettings> guardY(NoiseSettings dimensionType) {
         if (dimensionType.minY() + dimensionType.height() > DimensionType.MAX_Y + 1) {
@@ -32,17 +_,25 @@
         }
     }
 
-    public static NoiseSettings create(int minY, int height) {
-        NoiseSettings noiseSettings = new NoiseSettings(minY, height);
+    public static NoiseSettings create(int minY, int height, int noiseSizeHorizontal, int noiseSizeVertical) {
+        NoiseSettings noiseSettings = new NoiseSettings(minY, height, noiseSizeHorizontal, noiseSizeVertical);
         guardY(noiseSettings).error().ifPresent(error -> {
             throw new IllegalStateException(error.message());
         });
         return noiseSettings;
     }
 
+    public int getCellHeight() {
+        return QuartPos.toBlock(this.noiseSizeVertical());
+    }
+
+    public int getCellWidth() {
+        return QuartPos.toBlock(this.noiseSizeHorizontal());
+    }
+
     public NoiseSettings clampToHeightAccessor(LevelHeightAccessor heightAccessor) {
         int newMinY = Math.max(this.minY, heightAccessor.getMinY());
         int newHeight = Math.min(this.minY + this.height, heightAccessor.getMaxY() + 1) - newMinY;
-        return new NoiseSettings(newMinY, newHeight);
+        return new NoiseSettings(newMinY, newHeight, this.noiseSizeHorizontal, this.noiseSizeVertical);
     }
 }
