--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -47,7 +_,7 @@
 import org.apache.commons.lang3.mutable.MutableObject;
 import org.jspecify.annotations.Nullable;
 
-public class NoiseBasedChunkGenerator extends ChunkGenerator {
+public final class NoiseBasedChunkGenerator extends ChunkGenerator {
     public static final MapCodec<NoiseBasedChunkGenerator> CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(
                 BiomeSource.CODEC.fieldOf("biome_source").forGetter(g -> g.biomeSource),
@@ -56,7 +_,7 @@
             .apply(i, i.stable(NoiseBasedChunkGenerator::new))
     );
     private static final BlockState AIR = Blocks.AIR.defaultBlockState();
-    public final Holder<NoiseGeneratorSettings> settings;
+    private final Holder<NoiseGeneratorSettings> settings;
     private final Supplier<Aquifer.FluidPicker> globalFluidPicker;
 
     public NoiseBasedChunkGenerator(BiomeSource biomeSource, Holder<NoiseGeneratorSettings> settings) {
@@ -74,7 +_,10 @@
             if (SharedConstants.DEBUG_DISABLE_FLUID_GENERATION) {
                 return emptyStatus;
             } else {
-                return y < Math.min(-54, seaLevel) ? lavaStatus : seaStatus;
+                // C2ME: Optimize world gen math
+                final int min = Math.min(-54, seaLevel);
+                return y < min ? lavaStatus : seaStatus;
+                // C2ME: Optimize world gen math
             }
         };
     }
@@ -84,7 +_,7 @@
         return CompletableFuture.supplyAsync(() -> {
             this.doCreateBiomes(blender, randomState, structureManager, protoChunk);
             return protoChunk;
-        }, Util.backgroundExecutor().forName("init_biomes"));
+        }, Runnable::run); // Paper - rewrite chunk system
     }
 
     private void doCreateBiomes(Blender blender, RandomState randomState, StructureManager structureManager, ChunkAccess protoChunk) {
@@ -191,7 +_,7 @@
         );
     }
 
-    protected OptionalInt iterateNoiseColumn(
+    private OptionalInt iterateNoiseColumn(
         LevelHeightAccessor heightAccessor,
         RandomState randomState,
         int blockX,
@@ -269,7 +_,7 @@
     @Override
     public void buildSurface(WorldGenRegion region, StructureManager structureManager, RandomState randomState, ChunkAccess protoChunk) {
         if (!SharedConstants.debugVoidTerrain(protoChunk.getPos()) && !SharedConstants.DEBUG_DISABLE_SURFACE) {
-            WorldGenerationContext context = new WorldGenerationContext(this, region);
+            WorldGenerationContext context = new WorldGenerationContext(this, region, region.getMinecraftWorld()); // Paper - Flat bedrock generator settings
             Set<Holder<Biome>> possibleBiomes = collectPossibleBiomes(region, 1);
             this.buildSurface(protoChunk, context, randomState, structureManager, region.getBiomeManager(), Blender.of(region), possibleBiomes);
         }
@@ -440,7 +_,7 @@
                     section.release();
                 }
             }
-        }, Util.backgroundExecutor().forName("wgen_fill_noise"));
+        }, Runnable::run); // Paper - rewrite chunk system
     }
 
     private ChunkAccess doFill(
@@ -523,7 +_,7 @@
 
     private BlockState debugPreliminarySurfaceLevel(NoiseChunk noiseChunk, int posX, int posY, int posZ, BlockState state) {
         if (SharedConstants.DEBUG_AQUIFERS && posZ >= 0 && posZ % 4 == 0) {
-            int preliminarySurfaceLevel = noiseChunk.computePreliminarySurfaceLevel(posX, posZ);
+            int preliminarySurfaceLevel = noiseChunk.preliminarySurfaceLevel(posX, posZ);
             int adjustedSurfaceLevel = preliminarySurfaceLevel + 8;
             if (posY == adjustedSurfaceLevel) {
                 state = adjustedSurfaceLevel < this.getSeaLevel() ? Blocks.SLIME_BLOCK.defaultBlockState() : Blocks.HONEY_BLOCK.defaultBlockState();
@@ -552,10 +_,10 @@
     public void spawnOriginalMobs(WorldGenRegion worldGenRegion) {
         if (!this.settings.value().disableMobGeneration()) {
             ChunkPos center = worldGenRegion.getCenter();
-            BlockPos sourcePos = center.getWorldPosition().atY(worldGenRegion.getMaxY());
+            Holder<Biome> biome = worldGenRegion.getBiome(center.getWorldPosition().atY(worldGenRegion.getMaxY()));
             WorldgenRandom random = new WorldgenRandom(new LegacyRandomSource(RandomSupport.generateUniqueSeed()));
             random.setDecorationSeed(worldGenRegion.getSeed(), center.getMinBlockX(), center.getMinBlockZ());
-            NaturalSpawner.spawnMobsForChunkGeneration(worldGenRegion, sourcePos, center, random);
+            NaturalSpawner.spawnMobsForChunkGeneration(worldGenRegion, biome, center, random);
         }
     }
 }
