--- a/net/minecraft/world/level/levelgen/NoiseRouterData.java
+++ b/net/minecraft/world/level/levelgen/NoiseRouterData.java
@@ -1,11 +_,10 @@
 package net.minecraft.world.level.levelgen;
 
-import it.unimi.dsi.fastutil.floats.FloatList;
+import it.unimi.dsi.fastutil.doubles.DoubleList;
 import java.util.List;
 import java.util.stream.Stream;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
-import net.minecraft.core.Vec3i;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.data.worldgen.BootstrapContext;
 import net.minecraft.data.worldgen.TerrainProvider;
@@ -14,31 +_,26 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.level.biome.OverworldBiomeBuilder;
 import net.minecraft.world.level.dimension.DimensionType;
-import net.minecraft.world.level.levelgen.densityfunction.DensityFunction;
-import net.minecraft.world.level.levelgen.densityfunction.DensityFunctions;
-import net.minecraft.world.level.levelgen.densityfunction.DistanceMetric;
-import net.minecraft.world.level.levelgen.densityfunction.op.SplineFunction;
-import net.minecraft.world.level.levelgen.material.rule.OreVeinRule;
 import net.minecraft.world.level.levelgen.synth.BlendedNoise;
 import net.minecraft.world.level.levelgen.synth.NormalNoise;
 
 public class NoiseRouterData {
     public static final float GLOBAL_OFFSET = -0.50375F;
     private static final float ORE_THICKNESS = 0.08F;
-    private static final float VEININESS_FREQUENCY = 1.5F;
-    private static final float NOODLE_SPACING_AND_STRAIGHTNESS = 1.5F;
-    private static final float SURFACE_DENSITY_THRESHOLD = 1.5625F;
-    private static final float CHEESE_NOISE_TARGET = -0.703125F;
-    public static final float NOISE_ZERO = 0.390625F;
+    private static final double VEININESS_FREQUENCY = 1.5;
+    private static final double NOODLE_SPACING_AND_STRAIGHTNESS = 1.5;
+    private static final double SURFACE_DENSITY_THRESHOLD = 1.5625;
+    private static final double CHEESE_NOISE_TARGET = -0.703125;
+    public static final double NOISE_ZERO = 0.390625;
     public static final int ISLAND_CHUNK_DISTANCE = 64;
     public static final long ISLAND_CHUNK_DISTANCE_SQR = 4096L;
     private static final int DENSITY_Y_ANCHOR_BOTTOM = -64;
     private static final int DENSITY_Y_ANCHOR_TOP = 320;
-    private static final float DENSITY_Y_BOTTOM = 1.5F;
-    private static final float DENSITY_Y_TOP = -1.5F;
+    private static final double DENSITY_Y_BOTTOM = 1.5;
+    private static final double DENSITY_Y_TOP = -1.5;
     private static final int OVERWORLD_BOTTOM_SLIDE_HEIGHT = 24;
-    private static final float BASE_DENSITY_MULTIPLIER = 4.0F;
-    private static final DensityFunction BLENDING_FACTOR = DensityFunctions.constant(10.0F);
+    private static final double BASE_DENSITY_MULTIPLIER = 4.0;
+    private static final DensityFunction BLENDING_FACTOR = DensityFunctions.constant(10.0);
     private static final DensityFunction BLENDING_JAGGEDNESS = DensityFunctions.zero();
     private static final ResourceKey<DensityFunction> ZERO = createKey("zero");
     private static final ResourceKey<DensityFunction> Y = createKey("y");
@@ -59,9 +_,7 @@
         createKey("overworld/jaggedness"),
         createKey("overworld/depth"),
         createKey("overworld/sloped_cheese"),
-        createKey("overworld/preliminary_surface_level"),
-        createKey("overworld/chunk_surface_level"),
-        createKey("overworld/final_density")
+        createKey("overworld/preliminary_surface_level")
     );
     public static final OverworldFunctionSet<ResourceKey<DensityFunction>> AMPLIFIED_OVERWORLD_FUNCTIONS = new OverworldFunctionSet<>(
         OVERWORLD_FUNCTIONS.temperature(),
@@ -73,9 +_,7 @@
         createKey("overworld_amplified/jaggedness"),
         createKey("overworld_amplified/depth"),
         createKey("overworld_amplified/sloped_cheese"),
-        createKey("overworld_amplified/preliminary_surface_level"),
-        createKey("overworld_amplified/chunk_surface_level"),
-        createKey("overworld_amplified/final_density")
+        createKey("overworld_amplified/preliminary_surface_level")
     );
     public static final OverworldFunctionSet<ResourceKey<DensityFunction>> LARGE_OVERWORLD_FUNCTIONS = new OverworldFunctionSet<>(
         createKey("overworld_large_biomes/temperature"),
@@ -87,11 +_,8 @@
         createKey("overworld_large_biomes/jaggedness"),
         createKey("overworld_large_biomes/depth"),
         createKey("overworld_large_biomes/sloped_cheese"),
-        createKey("overworld_large_biomes/preliminary_surface_level"),
-        createKey("overworld_large_biomes/chunk_surface_level"),
-        createKey("overworld_large_biomes/final_density")
+        createKey("overworld_large_biomes/preliminary_surface_level")
     );
-    private static final ResourceKey<DensityFunction> END_ISLANDS = createKey("end/islands");
     private static final ResourceKey<DensityFunction> SLOPED_CHEESE_END = createKey("end/sloped_cheese");
     private static final ResourceKey<DensityFunction> SPAGHETTI_ROUGHNESS_FUNCTION = createKey("overworld/caves/spaghetti_roughness_function");
     private static final ResourceKey<DensityFunction> ENTRANCES = createKey("overworld/caves/entrances");
@@ -101,28 +_,32 @@
     private static final ResourceKey<DensityFunction> SPAGHETTI_2D = createKey("overworld/caves/spaghetti_2d");
     private static final ResourceKey<DensityFunction> ORE_VEIN_MASK = createKey("overworld/ore_vein/mask");
     private static final ResourceKey<DensityFunction> ORE_VEIN_TOGGLE = createKey("overworld/ore_vein/toggle");
-    public static final ResourceKey<DensityFunction> ORE_VEIN_RICHNESS = createKey("overworld/ore_vein/richness");
-    public static final ResourceKey<DensityFunction> ORE_VEIN_COPPER_DENSITY = createKey("overworld/ore_vein/copper_density");
-    public static final ResourceKey<DensityFunction> ORE_VEIN_IRON_DENSITY = createKey("overworld/ore_vein/iron_density");
-    public static final ResourceKey<DensityFunction> ORE_VEIN_GAP = createKey("overworld/ore_vein/gap");
+    private static final ResourceKey<DensityFunction> ORE_VEIN_RICHNESS = createKey("overworld/ore_vein/richness");
+    private static final ResourceKey<DensityFunction> ORE_VEIN_COPPER_DENSITY = createKey("overworld/ore_vein/copper_density");
+    private static final ResourceKey<DensityFunction> ORE_VEIN_IRON_DENSITY = createKey("overworld/ore_vein/iron_density");
+    private static final ResourceKey<DensityFunction> ORE_VEIN_GAP = createKey("overworld/ore_vein/gap");
 
     private static ResourceKey<DensityFunction> createKey(String name) {
         return ResourceKey.create(Registries.DENSITY_FUNCTION, Identifier.withDefaultNamespace(name));
     }
 
     public static Holder<? extends DensityFunction> bootstrap(BootstrapContext<DensityFunction> context) {
-        HolderGetter<NormalNoise> noises = context.lookup(Registries.NOISE);
+        HolderGetter<NormalNoise.NoiseParameters> noises = context.lookup(Registries.NOISE);
         HolderGetter<DensityFunction> functions = context.lookup(Registries.DENSITY_FUNCTION);
         context.register(ZERO, DensityFunctions.zero());
         int belowBottom = DimensionType.MIN_Y * 2;
         int aboveTop = DimensionType.MAX_Y * 2;
         context.register(Y, DensityFunctions.yClampedGradient(belowBottom, aboveTop, belowBottom, aboveTop));
         registerOreVeins(context);
-        DensityFunction shiftX = registerAndWrap(context, SHIFT_X, DensityFunctions.cache(DensityFunctions.shiftA(noises.getOrThrow(Noises.SHIFT))));
-        DensityFunction shiftZ = registerAndWrap(context, SHIFT_Z, DensityFunctions.cache(DensityFunctions.shiftB(noises.getOrThrow(Noises.SHIFT))));
-        context.register(BASE_3D_NOISE_OVERWORLD, new BlendedNoise(0.25, 0.125, 80.0, 160.0, 8.0));
-        context.register(BASE_3D_NOISE_NETHER, new BlendedNoise(0.25, 0.375, 80.0, 60.0, 8.0));
-        context.register(BASE_3D_NOISE_END, new BlendedNoise(0.25, 0.25, 80.0, 160.0, 4.0));
+        DensityFunction shiftX = registerAndWrap(
+            context, SHIFT_X, DensityFunctions.flatCache(DensityFunctions.cache2d(DensityFunctions.shiftA(noises.getOrThrow(Noises.SHIFT))))
+        );
+        DensityFunction shiftZ = registerAndWrap(
+            context, SHIFT_Z, DensityFunctions.flatCache(DensityFunctions.cache2d(DensityFunctions.shiftB(noises.getOrThrow(Noises.SHIFT))))
+        );
+        context.register(BASE_3D_NOISE_OVERWORLD, BlendedNoise.createUnseeded(0.25, 0.125, 80.0, 160.0, 8.0));
+        context.register(BASE_3D_NOISE_NETHER, BlendedNoise.createUnseeded(0.25, 0.375, 80.0, 60.0, 8.0));
+        context.register(BASE_3D_NOISE_END, BlendedNoise.createUnseeded(0.25, 0.25, 80.0, 160.0, 4.0));
         registerAndWrap(
             context, OVERWORLD_FUNCTIONS.temperature(), DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.TEMPERATURE))
         );
@@ -138,37 +_,36 @@
         DensityFunction continents = registerAndWrap(
             context,
             OVERWORLD_FUNCTIONS.continents(),
-            DensityFunctions.cache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.CONTINENTALNESS)))
+            DensityFunctions.flatCache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.CONTINENTALNESS)))
         );
         DensityFunction erosion = registerAndWrap(
             context,
             OVERWORLD_FUNCTIONS.erosion(),
-            DensityFunctions.cache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.EROSION)))
+            DensityFunctions.flatCache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.EROSION)))
         );
         DensityFunction ridge = registerAndWrap(
-            context, RIDGES, DensityFunctions.cache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.RIDGE)))
+            context, RIDGES, DensityFunctions.flatCache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.RIDGE)))
         );
         context.register(RIDGES_FOLDED, peaksAndValleys(ridge));
         DensityFunction jaggedNoise = DensityFunctions.noise(noises.getOrThrow(Noises.JAGGED), 1500.0, 0.0);
-        registerTerrainNoises(context, functions, noises, jaggedNoise, continents, erosion, OVERWORLD_FUNCTIONS, false);
+        registerTerrainNoises(context, functions, jaggedNoise, continents, erosion, OVERWORLD_FUNCTIONS, false);
         DensityFunction continentsLarge = registerAndWrap(
             context,
             LARGE_OVERWORLD_FUNCTIONS.continents(),
-            DensityFunctions.cache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.CONTINENTALNESS_LARGE)))
+            DensityFunctions.flatCache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.CONTINENTALNESS_LARGE)))
         );
         DensityFunction erosionLarge = registerAndWrap(
             context,
             LARGE_OVERWORLD_FUNCTIONS.erosion(),
-            DensityFunctions.cache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.EROSION_LARGE)))
+            DensityFunctions.flatCache(DensityFunctions.shiftedNoise2d(shiftX, shiftZ, 0.25, noises.getOrThrow(Noises.EROSION_LARGE)))
         );
-        registerTerrainNoises(context, functions, noises, jaggedNoise, continentsLarge, erosionLarge, LARGE_OVERWORLD_FUNCTIONS, false);
-        registerTerrainNoises(context, functions, noises, jaggedNoise, continents, erosion, AMPLIFIED_OVERWORLD_FUNCTIONS, true);
-        DensityFunction endIslands = registerAndWrap(context, END_ISLANDS, createEndIslands());
-        context.register(SLOPED_CHEESE_END, DensityFunctions.add(endIslands, getFunction(functions, BASE_3D_NOISE_END)));
+        registerTerrainNoises(context, functions, jaggedNoise, continentsLarge, erosionLarge, LARGE_OVERWORLD_FUNCTIONS, false);
+        registerTerrainNoises(context, functions, jaggedNoise, continents, erosion, AMPLIFIED_OVERWORLD_FUNCTIONS, true);
+        context.register(SLOPED_CHEESE_END, DensityFunctions.add(DensityFunctions.endIslands(0L), getFunction(functions, BASE_3D_NOISE_END)));
         context.register(SPAGHETTI_ROUGHNESS_FUNCTION, spaghettiRoughnessFunction(noises));
         context.register(
             SPAGHETTI_2D_THICKNESS_MODULATOR,
-            DensityFunctions.cache(DensityFunctions.mappedNoise(noises.getOrThrow(Noises.SPAGHETTI_2D_THICKNESS), 2.0, 1.0, -0.6F, -1.3F))
+            DensityFunctions.cacheOnce(DensityFunctions.mappedNoise(noises.getOrThrow(Noises.SPAGHETTI_2D_THICKNESS), 2.0, 1.0, -0.6, -1.3))
         );
         context.register(SPAGHETTI_2D, spaghetti2D(functions, noises));
         context.register(ENTRANCES, entrances(functions, noises));
@@ -176,29 +_,19 @@
         return context.register(PILLARS, pillars(noises));
     }
 
-    private static DensityFunction createEndIslands() {
-        DensityFunction distanceToMainIsland = DensityFunctions.distanceToPoint(Vec3i.ZERO, DistanceMetric.EUCLIDEAN);
-        DensityFunction mainIsland = DensityFunctions.sliceY(
-            DensityFunctions.constant(100.0F).sub(distanceToMainIsland).clamp(-100.0F, 80.0F).sub(8.0F).mul(0.0078125F), 0
-        );
-        DensityFunction outerIslands = DensityFunctions.endOuterIslands();
-        return DensityFunctions.cache(DensityFunctions.max(mainIsland, outerIslands));
-    }
-
     private static void registerTerrainNoises(
         BootstrapContext<DensityFunction> context,
         HolderGetter<DensityFunction> functions,
-        HolderGetter<NormalNoise> noises,
         DensityFunction jaggedNoise,
         DensityFunction continentsFunction,
         DensityFunction erosionFunction,
         OverworldFunctionSet<ResourceKey<DensityFunction>> names,
         boolean amplified
     ) {
-        SplineFunction.Coordinate continents = new SplineFunction.Coordinate(continentsFunction);
-        SplineFunction.Coordinate erosion = new SplineFunction.Coordinate(erosionFunction);
-        SplineFunction.Coordinate weirdness = new SplineFunction.Coordinate(getFunction(functions, RIDGES));
-        SplineFunction.Coordinate ridges = new SplineFunction.Coordinate(getFunction(functions, RIDGES_FOLDED));
+        DensityFunctions.Spline.Coordinate continents = new DensityFunctions.Spline.Coordinate(continentsFunction);
+        DensityFunctions.Spline.Coordinate erosion = new DensityFunctions.Spline.Coordinate(erosionFunction);
+        DensityFunctions.Spline.Coordinate weirdness = new DensityFunctions.Spline.Coordinate(getFunction(functions, RIDGES));
+        DensityFunctions.Spline.Coordinate ridges = new DensityFunctions.Spline.Coordinate(getFunction(functions, RIDGES_FOLDED));
         DensityFunction offset = registerAndWrap(
             context,
             names.offset(),
@@ -222,41 +_,30 @@
                 DensityFunctions.spline(TerrainProvider.overworldJaggedness(continents, erosion, weirdness, ridges, amplified)), BLENDING_JAGGEDNESS
             )
         );
-        DensityFunction jaggedness = DensityFunctions.cache(DensityFunctions.mul(unscaledJaggedness, jaggedNoise.halfNegative()));
+        DensityFunction jaggedness = DensityFunctions.flatCache(DensityFunctions.mul(unscaledJaggedness, jaggedNoise.halfNegative()));
         DensityFunction initialDensity = noiseGradientDensity(factor, DensityFunctions.add(depth, jaggedness));
-        DensityFunction slopedCheese = registerAndWrap(
-            context, names.slopedCheese(), DensityFunctions.cache(DensityFunctions.add(initialDensity, getFunction(functions, BASE_3D_NOISE_OVERWORLD)))
-        );
-        DensityFunction surfaceLevel = registerAndWrap(context, names.preliminarySurfaceLevel(), preliminarySurfaceLevel(offset, factor, amplified));
-        context.register(names.chunkSurfaceLevel(), DensityFunctions.interpolated(surfaceLevel, 16, 1));
-        DensityFunction surfaceWithEntrances = DensityFunctions.min(slopedCheese, getFunction(functions, ENTRANCES).mul(5.0F));
-        DensityFunction caves = DensityFunctions.rangeChoice(
-            slopedCheese, -1000000.0F, 1.5625F, surfaceWithEntrances, underground(functions, noises, slopedCheese)
-        );
-        context.register(
-            names.finalDensity(),
-            DensityFunctions.add(
-                DensityFunctions.min(postProcess(slideOverworld(amplified, caves), 4, 8), getFunction(functions, NOODLE)), DensityFunctions.beardifier()
-            )
-        );
+        context.register(names.slopedCheese(), DensityFunctions.add(initialDensity, getFunction(functions, BASE_3D_NOISE_OVERWORLD)));
+        context.register(names.preliminarySurfaceLevel(), preliminarySurfaceLevel(offset, factor, amplified));
     }
 
     private static DensityFunction offsetToDepth(DensityFunction offset) {
-        return DensityFunctions.add(DensityFunctions.yClampedGradient(-64, 320, 1.5F, -1.5F), offset);
+        return DensityFunctions.add(DensityFunctions.yClampedGradient(-64, 320, 1.5, -1.5), offset);
     }
 
     private static DensityFunction registerAndWrap(BootstrapContext<DensityFunction> context, ResourceKey<DensityFunction> name, DensityFunction value) {
         return new DensityFunctions.HolderHolder(context.register(name, value));
     }
 
-    public static DensityFunction getFunction(HolderGetter<DensityFunction> functions, ResourceKey<DensityFunction> name) {
+    private static DensityFunction getFunction(HolderGetter<DensityFunction> functions, ResourceKey<DensityFunction> name) {
         return new DensityFunctions.HolderHolder(functions.getOrThrow(name));
     }
 
-    public static DensityFunction peaksAndValleys(DensityFunction weirdness) {
+    private static DensityFunction peaksAndValleys(DensityFunction weirdness) {
         return DensityFunctions.mul(
-            DensityFunctions.add(DensityFunctions.add(weirdness.abs(), DensityFunctions.constant(-0.6666667F)).abs(), DensityFunctions.constant(-0.33333334F)),
-            DensityFunctions.constant(-3.0F)
+            DensityFunctions.add(
+                DensityFunctions.add(weirdness.abs(), DensityFunctions.constant(-0.6666666666666666)).abs(), DensityFunctions.constant(-0.3333333333333333)
+            ),
+            DensityFunctions.constant(-3.0)
         );
     }
 
@@ -264,17 +_,17 @@
         return TerrainProvider.peaksAndValleys(weirdness);
     }
 
-    private static DensityFunction spaghettiRoughnessFunction(HolderGetter<NormalNoise> noises) {
+    private static DensityFunction spaghettiRoughnessFunction(HolderGetter<NormalNoise.NoiseParameters> noises) {
         DensityFunction spaghettiRoughnessNoise = DensityFunctions.noise(noises.getOrThrow(Noises.SPAGHETTI_ROUGHNESS));
-        DensityFunction spaghettiRoughnessModulator = DensityFunctions.mappedNoise(noises.getOrThrow(Noises.SPAGHETTI_ROUGHNESS_MODULATOR), 0.0F, -0.1F);
-        return DensityFunctions.cache(
-            DensityFunctions.mul(spaghettiRoughnessModulator, DensityFunctions.add(spaghettiRoughnessNoise.abs(), DensityFunctions.constant(-0.4F)))
+        DensityFunction spaghettiRoughnessModulator = DensityFunctions.mappedNoise(noises.getOrThrow(Noises.SPAGHETTI_ROUGHNESS_MODULATOR), 0.0, -0.1);
+        return DensityFunctions.cacheOnce(
+            DensityFunctions.mul(spaghettiRoughnessModulator, DensityFunctions.add(spaghettiRoughnessNoise.abs(), DensityFunctions.constant(-0.4)))
         );
     }
 
-    private static DensityFunction entrances(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise> noises) {
-        DensityFunction spaghetti3DRarityModulator = DensityFunctions.cache(DensityFunctions.noise(noises.getOrThrow(Noises.SPAGHETTI_3D_RARITY), 2.0, 1.0));
-        DensityFunction spaghetti3DThicknessModulator = DensityFunctions.mappedNoise(noises.getOrThrow(Noises.SPAGHETTI_3D_THICKNESS), -0.065F, -0.088F);
+    private static DensityFunction entrances(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise.NoiseParameters> noises) {
+        DensityFunction spaghetti3DRarityModulator = DensityFunctions.cacheOnce(DensityFunctions.noise(noises.getOrThrow(Noises.SPAGHETTI_3D_RARITY), 2.0, 1.0));
+        DensityFunction spaghetti3DThicknessModulator = DensityFunctions.mappedNoise(noises.getOrThrow(Noises.SPAGHETTI_3D_THICKNESS), -0.065, -0.088);
         DensityFunction spaghetti3DCave1 = NoiseRouterData.QuantizedSpaghettiRarity.wrapRarity3d(
             spaghetti3DRarityModulator, noises.getOrThrow(Noises.SPAGHETTI_3D_1)
         );
@@ -282,93 +_,102 @@
             spaghetti3DRarityModulator, noises.getOrThrow(Noises.SPAGHETTI_3D_2)
         );
         DensityFunction spaghetti3DFunction = DensityFunctions.add(DensityFunctions.max(spaghetti3DCave1, spaghetti3DCave2), spaghetti3DThicknessModulator)
-            .clamp(-1.0F, 1.0F);
+            .clamp(-1.0, 1.0);
         DensityFunction spaghettiRoughnessFunction = getFunction(functions, SPAGHETTI_ROUGHNESS_FUNCTION);
         DensityFunction bigEntranceNoiseSource = DensityFunctions.noise(noises.getOrThrow(Noises.CAVE_ENTRANCE), 0.75, 0.5);
-        DensityFunction bigEntrancesFunction = DensityFunctions.add(bigEntranceNoiseSource.add(0.37F), DensityFunctions.yClampedGradient(-10, 30, 0.3F, 0.0F));
-        return DensityFunctions.cache(DensityFunctions.min(bigEntrancesFunction, DensityFunctions.add(spaghettiRoughnessFunction, spaghetti3DFunction)));
+        DensityFunction bigEntrancesFunction = DensityFunctions.add(bigEntranceNoiseSource.add(0.37), DensityFunctions.yClampedGradient(-10, 30, 0.3, 0.0));
+        return DensityFunctions.cacheOnce(DensityFunctions.min(bigEntrancesFunction, DensityFunctions.add(spaghettiRoughnessFunction, spaghetti3DFunction)));
     }
 
-    private static DensityFunction noodle(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise> noises) {
+    private static DensityFunction noodle(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise.NoiseParameters> noises) {
         DensityFunction y = getFunction(functions, Y);
         int minBlockY = -64;
         int noodleMinY = -60;
         int noodleMaxY = 320;
-        int cellSizeXz = 4;
-        int cellSizeY = 8;
-        DensityFunction noodleToggle = yLimitedInterpolatable(y, DensityFunctions.noise(noises.getOrThrow(Noises.NOODLE), 1.0, 1.0), -60, 320, -1, 4, 8);
+        DensityFunction noodleToggle = yLimitedInterpolatable(y, DensityFunctions.noise(noises.getOrThrow(Noises.NOODLE), 1.0, 1.0), -60, 320, -1);
         DensityFunction noodleThickness = yLimitedInterpolatable(
-            y, DensityFunctions.mappedNoise(noises.getOrThrow(Noises.NOODLE_THICKNESS), 1.0, 1.0, -0.05F, -0.1F), -60, 320, 0, 4, 8
+            y, DensityFunctions.mappedNoise(noises.getOrThrow(Noises.NOODLE_THICKNESS), 1.0, 1.0, -0.05, -0.1), -60, 320, 0
         );
         double noodleRidgeFrequency = 2.6666666666666665;
         DensityFunction noodleRidgeA = yLimitedInterpolatable(
-            y, DensityFunctions.noise(noises.getOrThrow(Noises.NOODLE_RIDGE_A), 2.6666666666666665, 2.6666666666666665), -60, 320, 0, 4, 8
+            y, DensityFunctions.noise(noises.getOrThrow(Noises.NOODLE_RIDGE_A), 2.6666666666666665, 2.6666666666666665), -60, 320, 0
         );
         DensityFunction noodleRidgeB = yLimitedInterpolatable(
-            y, DensityFunctions.noise(noises.getOrThrow(Noises.NOODLE_RIDGE_B), 2.6666666666666665, 2.6666666666666665), -60, 320, 0, 4, 8
-        );
-        DensityFunction noodleRidged = DensityFunctions.max(noodleRidgeA.abs(), noodleRidgeB.abs()).mul(1.5F);
-        return DensityFunctions.rangeChoice(
-            noodleToggle, -1000000.0F, 0.0F, DensityFunctions.constant(64.0F), DensityFunctions.add(noodleThickness, noodleRidged)
-        );
+            y, DensityFunctions.noise(noises.getOrThrow(Noises.NOODLE_RIDGE_B), 2.6666666666666665, 2.6666666666666665), -60, 320, 0
+        );
+        DensityFunction noodleRidged = DensityFunctions.max(noodleRidgeA.abs(), noodleRidgeB.abs()).mul(1.5);
+        return DensityFunctions.rangeChoice(noodleToggle, -1000000.0, 0.0, DensityFunctions.constant(64.0), DensityFunctions.add(noodleThickness, noodleRidged));
     }
 
-    private static DensityFunction pillars(HolderGetter<NormalNoise> noises) {
+    private static DensityFunction pillars(HolderGetter<NormalNoise.NoiseParameters> noises) {
         double xzFrequency = 25.0;
         double yFrequency = 0.3;
         DensityFunction pillarNoiseSource = DensityFunctions.noise(noises.getOrThrow(Noises.PILLAR), 25.0, 0.3);
-        DensityFunction pillarRarenessModulator = DensityFunctions.mappedNoise(noises.getOrThrow(Noises.PILLAR_RARENESS), 0.0F, -2.0F);
-        DensityFunction pillarThicknessModulator = DensityFunctions.mappedNoise(noises.getOrThrow(Noises.PILLAR_THICKNESS), 0.0F, 1.1F);
-        DensityFunction pillarsWithRareness = DensityFunctions.add(pillarNoiseSource.mul(2.0F), pillarRarenessModulator);
-        return DensityFunctions.cache(DensityFunctions.mul(pillarsWithRareness, pillarThicknessModulator.cube()));
+        DensityFunction pillarRarenessModulator = DensityFunctions.mappedNoise(noises.getOrThrow(Noises.PILLAR_RARENESS), 0.0, -2.0);
+        DensityFunction pillarThicknessModulator = DensityFunctions.mappedNoise(noises.getOrThrow(Noises.PILLAR_THICKNESS), 0.0, 1.1);
+        DensityFunction pillarsWithRareness = DensityFunctions.add(pillarNoiseSource.mul(2.0), pillarRarenessModulator);
+        return DensityFunctions.cacheOnce(DensityFunctions.mul(pillarsWithRareness, pillarThicknessModulator.cube()));
     }
 
-    private static DensityFunction spaghetti2D(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise> noises) {
+    private static DensityFunction spaghetti2D(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise.NoiseParameters> noises) {
         DensityFunction spaghetti2DRarityModulator = DensityFunctions.noise(noises.getOrThrow(Noises.SPAGHETTI_2D_MODULATOR), 2.0, 1.0);
         DensityFunction spaghetti2DCave = NoiseRouterData.QuantizedSpaghettiRarity.wrapRarity2d(
             spaghetti2DRarityModulator, noises.getOrThrow(Noises.SPAGHETTI_2D)
         );
         DensityFunction spaghetti2DElevationModulator = DensityFunctions.mappedNoise(
-            noises.getOrThrow(Noises.SPAGHETTI_2D_ELEVATION), 0.0, Math.floorDiv(-64, 8), 8.0F
+            noises.getOrThrow(Noises.SPAGHETTI_2D_ELEVATION), 0.0, Math.floorDiv(-64, 8), 8.0
         );
         DensityFunction spaghetti2DThicknessModulator = getFunction(functions, SPAGHETTI_2D_THICKNESS_MODULATOR);
         DensityFunction slopedSpaghetti = DensityFunctions.add(
-                DensityFunctions.cache(spaghetti2DElevationModulator), DensityFunctions.yClampedGradient(-64, 320, 8.0F, -40.0F)
+                DensityFunctions.flatCache(spaghetti2DElevationModulator), DensityFunctions.yClampedGradient(-64, 320, 8.0, -40.0)
             )
             .abs();
         DensityFunction layerRidged = DensityFunctions.add(slopedSpaghetti, spaghetti2DThicknessModulator).cube();
-        float ridgeOffset = 0.083F;
-        DensityFunction caveNoise = DensityFunctions.add(spaghetti2DCave, spaghetti2DThicknessModulator.mul(0.083F));
-        return DensityFunctions.max(caveNoise, layerRidged).clamp(-1.0F, 1.0F);
+        double ridgeOffset = 0.083;
+        DensityFunction caveNoise = DensityFunctions.add(spaghetti2DCave, spaghetti2DThicknessModulator.mul(0.083));
+        return DensityFunctions.max(caveNoise, layerRidged).clamp(-1.0, 1.0);
     }
 
-    private static DensityFunction underground(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise> noises, DensityFunction slopedCheese) {
+    private static DensityFunction underground(
+        HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise.NoiseParameters> noises, DensityFunction slopedCheese
+    ) {
         DensityFunction spaghetti2DFunction = getFunction(functions, SPAGHETTI_2D);
         DensityFunction spaghettiRoughnessFunction = getFunction(functions, SPAGHETTI_ROUGHNESS_FUNCTION);
         DensityFunction layerNoiseSource = DensityFunctions.noise(noises.getOrThrow(Noises.CAVE_LAYER), 8.0);
-        DensityFunction layerizedCavernsFunction = layerNoiseSource.square().mul(4.0F);
+        DensityFunction layerizedCavernsFunction = layerNoiseSource.square().mul(4.0);
         DensityFunction cheese = DensityFunctions.noise(noises.getOrThrow(Noises.CAVE_CHEESE), 0.6666666666666666);
-        DensityFunction solidifedCheeseWithTopSlide = DensityFunctions.add(
-            cheese.add(0.27F).clamp(-1.0F, 1.0F), slopedCheese.mul(-0.64F).add(1.5F).clamp(0.0F, 0.5F)
-        );
+        DensityFunction solidifedCheeseWithTopSlide = DensityFunctions.add(cheese.add(0.27).clamp(-1.0, 1.0), slopedCheese.mul(-0.64).add(1.5).clamp(0.0, 0.5));
         DensityFunction baseCaveDensity = DensityFunctions.add(layerizedCavernsFunction, solidifedCheeseWithTopSlide);
         DensityFunction undergroundSubtractions = DensityFunctions.min(
             DensityFunctions.min(baseCaveDensity, getFunction(functions, ENTRANCES)), DensityFunctions.add(spaghetti2DFunction, spaghettiRoughnessFunction)
         );
         DensityFunction pillarsWithoutCutoff = getFunction(functions, PILLARS);
         DensityFunction pillars = DensityFunctions.rangeChoice(
-            pillarsWithoutCutoff, -1000000.0F, 0.03F, DensityFunctions.constant(-1000000.0F), pillarsWithoutCutoff
+            pillarsWithoutCutoff, -1000000.0, 0.03, DensityFunctions.constant(-1000000.0), pillarsWithoutCutoff
         );
         return DensityFunctions.max(undergroundSubtractions, pillars);
     }
 
-    private static DensityFunction postProcess(DensityFunction slide, int cellSizeXz, int cellSizeY) {
+    private static DensityFunction postProcess(DensityFunction slide) {
         DensityFunction blended = DensityFunctions.blendDensity(slide);
-        return DensityFunctions.interpolated(DensityFunctions.mul(blended, DensityFunctions.constant(0.64F)), cellSizeXz, cellSizeY).squeeze();
+        return DensityFunctions.interpolated(DensityFunctions.mul(blended, DensityFunctions.constant(0.64))).squeeze();
     }
 
-    protected static NoiseRouter overworld(HolderGetter<DensityFunction> functions, OverworldFunctionSet<ResourceKey<DensityFunction>> functionNames) {
+    protected static NoiseRouter overworld(
+        HolderGetter<DensityFunction> functions,
+        HolderGetter<NormalNoise.NoiseParameters> noises,
+        OverworldFunctionSet<ResourceKey<DensityFunction>> functionNames,
+        boolean amplified
+    ) {
         OverworldFunctionSet<DensityFunction> functionSet = functionNames.map(key -> getFunction(functions, key));
+        DensityFunction slopedCheese = DensityFunctions.cacheOnce(functionSet.slopedCheese());
+        DensityFunction surfaceWithEntrances = DensityFunctions.min(slopedCheese, getFunction(functions, ENTRANCES).mul(5.0));
+        DensityFunction caves = DensityFunctions.rangeChoice(
+            slopedCheese, -1000000.0, 1.5625, surfaceWithEntrances, underground(functions, noises, slopedCheese)
+        );
+        DensityFunction fullNoise = DensityFunctions.add(
+            DensityFunctions.min(postProcess(slideOverworld(amplified, caves)), getFunction(functions, NOODLE)), DensityFunctions.beardifier()
+        );
         return new NoiseRouter(
             functionSet.temperature(),
             functionSet.vegetation(),
@@ -376,13 +_,13 @@
             functionSet.erosion(),
             functionSet.depth(),
             getFunction(functions, RIDGES),
-            functionSet.chunkSurfaceLevel(),
-            functionSet.finalDensity()
+            functionSet.preliminarySurfaceLevel(),
+            fullNoise
         );
     }
 
     protected static Aquifer.Config overworldAquifers(
-        HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise> noises, OverworldFunctionSet<ResourceKey<DensityFunction>> names
+        HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise.NoiseParameters> noises, OverworldFunctionSet<ResourceKey<DensityFunction>> names
     ) {
         DensityFunction barrierNoise = DensityFunctions.noise(noises.getOrThrow(Noises.AQUIFER_BARRIER), 0.5);
         DensityFunction fluidLevelFloodednessNoise = DensityFunctions.noise(noises.getOrThrow(Noises.AQUIFER_FLUID_LEVEL_FLOODEDNESS), 0.67);
@@ -396,49 +_,54 @@
 
     private static void registerOreVeins(BootstrapContext<DensityFunction> context) {
         HolderGetter<DensityFunction> functions = context.lookup(Registries.DENSITY_FUNCTION);
-        HolderGetter<NormalNoise> noises = context.lookup(Registries.NOISE);
+        HolderGetter<NormalNoise.NoiseParameters> noises = context.lookup(Registries.NOISE);
         DensityFunction y = getFunction(functions, Y);
-        int veinMinY = Stream.of(OreVeinRule.VeinType.values()).mapToInt(t -> t.minY).min().orElseThrow();
-        int veinMaxY = Stream.of(OreVeinRule.VeinType.values()).mapToInt(t -> t.maxY).max().orElseThrow();
-        int interpolatedCellHeight = 8;
-        int interpolatedVeinMinY = Mth.floorDiv(veinMinY, 8) * 8;
-        int interpolatedVeinMaxY = (Mth.floorDiv(veinMaxY, 8) + 1) * 8;
+        int veinMinY = Stream.of(OreVeinifier.VeinType.values()).mapToInt(t -> t.minY).min().orElseThrow();
+        int veinMaxY = Stream.of(OreVeinifier.VeinType.values()).mapToInt(t -> t.maxY).max().orElseThrow();
+        int interpolatedCellHeight = NoiseSettings.OVERWORLD_NOISE_SETTINGS.getCellHeight();
+        int interpolatedVeinMinY = Mth.floorDiv(veinMinY, interpolatedCellHeight) * interpolatedCellHeight;
+        int interpolatedVeinMaxY = (Mth.floorDiv(veinMaxY, interpolatedCellHeight) + 1) * interpolatedCellHeight;
         DensityFunction veinToggle = registerAndWrap(
             context,
             ORE_VEIN_TOGGLE,
-            DensityFunctions.cache(
-                yLimitedInterpolatable(
-                    y, DensityFunctions.noise(noises.getOrThrow(Noises.ORE_VEININESS), 1.5, 1.5), interpolatedVeinMinY, interpolatedVeinMaxY, 0, 4, 8
-                )
-            )
+            yLimitedInterpolatable(y, DensityFunctions.noise(noises.getOrThrow(Noises.ORE_VEININESS), 1.5, 1.5), interpolatedVeinMinY, interpolatedVeinMaxY, 0)
         );
         context.register(ORE_VEIN_RICHNESS, DensityFunctions.clampedMap(veinToggle.abs(), 0.4F, 0.6F, 0.1F, 0.3F));
         DensityFunction veinMask = registerAndWrap(
-            context, ORE_VEIN_MASK, DensityFunctions.cache(createBaseOreVeinMask(noises, y, veinToggle, interpolatedVeinMinY, interpolatedVeinMaxY))
+            context, ORE_VEIN_MASK, DensityFunctions.cacheOnce(createBaseOreVeinMask(noises, y, veinToggle, interpolatedVeinMinY, interpolatedVeinMaxY))
         );
-        context.register(ORE_VEIN_COPPER_DENSITY, createOreVeinDensity(OreVeinRule.VeinType.COPPER, y, veinToggle, veinMask, true));
-        context.register(ORE_VEIN_IRON_DENSITY, createOreVeinDensity(OreVeinRule.VeinType.IRON, y, veinToggle, veinMask, false));
+        context.register(ORE_VEIN_COPPER_DENSITY, createOreVeinDensity(OreVeinifier.VeinType.COPPER, y, veinToggle, veinMask, true));
+        context.register(ORE_VEIN_IRON_DENSITY, createOreVeinDensity(OreVeinifier.VeinType.IRON, y, veinToggle, veinMask, false));
         context.register(ORE_VEIN_GAP, DensityFunctions.constant(-0.3F).sub(DensityFunctions.noise(noises.getOrThrow(Noises.ORE_GAP))));
     }
 
-    private static DensityFunction createBaseOreVeinMask(HolderGetter<NormalNoise> noises, DensityFunction y, DensityFunction toggle, int minY, int maxY) {
+    protected static List<OreVeinifier> overworldOreVeins(HolderGetter<DensityFunction> functions) {
+        DensityFunction richness = getFunction(functions, ORE_VEIN_RICHNESS);
+        DensityFunction gap = getFunction(functions, ORE_VEIN_GAP);
+        return List.of(
+            OreVeinifier.VeinType.COPPER.create(getFunction(functions, ORE_VEIN_COPPER_DENSITY), richness, gap),
+            OreVeinifier.VeinType.IRON.create(getFunction(functions, ORE_VEIN_IRON_DENSITY), richness, gap)
+        );
+    }
+
+    private static DensityFunction createBaseOreVeinMask(
+        HolderGetter<NormalNoise.NoiseParameters> noises, DensityFunction y, DensityFunction toggle, int minY, int maxY
+    ) {
         float oreRidgeFrequency = 4.0F;
         int noVein = 1;
-        int cellSizeXz = 4;
-        int cellSizeY = 8;
-        DensityFunction veinA = yLimitedInterpolatable(y, DensityFunctions.noise(noises.getOrThrow(Noises.ORE_VEIN_A), 4.0, 4.0), minY, maxY, 1, 4, 8).abs();
-        DensityFunction veinB = yLimitedInterpolatable(y, DensityFunctions.noise(noises.getOrThrow(Noises.ORE_VEIN_B), 4.0, 4.0), minY, maxY, 1, 4, 8).abs();
+        DensityFunction veinA = yLimitedInterpolatable(y, DensityFunctions.noise(noises.getOrThrow(Noises.ORE_VEIN_A), 4.0, 4.0), minY, maxY, 1).abs();
+        DensityFunction veinB = yLimitedInterpolatable(y, DensityFunctions.noise(noises.getOrThrow(Noises.ORE_VEIN_B), 4.0, 4.0), minY, maxY, 1).abs();
         return DensityFunctions.rangeChoice(
-            toggle, -0.4F, 0.4F, DensityFunctions.constant(-1.0F), DensityFunctions.constant(0.08F).sub(DensityFunctions.max(veinA, veinB))
+            toggle, -0.4F, 0.4F, DensityFunctions.constant(-1.0), DensityFunctions.constant(0.08F).sub(DensityFunctions.max(veinA, veinB))
         );
     }
 
     private static DensityFunction createOreVeinDensity(
-        OreVeinRule.VeinType type, DensityFunction y, DensityFunction toggle, DensityFunction baseVeinMask, boolean whenTogglePositive
+        OreVeinifier.VeinType type, DensityFunction y, DensityFunction toggle, DensityFunction baseVeinMask, boolean whenTogglePositive
     ) {
-        DensityFunction noVein = DensityFunctions.constant(-1.0F);
-        DensityFunction distanceFromEdge = DensityFunctions.min(DensityFunctions.constant(type.maxY).sub(y), y.sub(type.minY));
-        DensityFunction edgeRoundoff = DensityFunctions.clampedMap(distanceFromEdge, 0.0F, 20.0F, -0.2F, 0.0F);
+        DensityFunction noVein = DensityFunctions.constant(-1.0);
+        DensityFunction distanceFromEdge = DensityFunctions.cacheOnce(DensityFunctions.min(DensityFunctions.constant(type.maxY).sub(y), y.sub(type.minY)));
+        DensityFunction edgeRoundoff = DensityFunctions.clampedMap(distanceFromEdge, 0.0, 20.0, -0.2, 0.0);
         DensityFunction veininess = whenTogglePositive ? toggle : toggle.negate();
         return DensityFunctions.rangeChoice(
             y,
@@ -446,9 +_,9 @@
             type.maxY,
             DensityFunctions.rangeChoice(
                 baseVeinMask,
-                0.0F,
-                1000000.0F,
-                DensityFunctions.rangeChoice(veininess.sub(0.4F).add(edgeRoundoff), 0.0F, 1000000.0F, DensityFunctions.constant(0.7F), noVein),
+                0.0,
+                1000000.0,
+                DensityFunctions.rangeChoice(veininess.sub(0.4F).add(edgeRoundoff), 0.0, 1000000.0, DensityFunctions.constant(0.7F), noVein),
                 noVein
             ),
             noVein
@@ -456,18 +_,18 @@
     }
 
     private static DensityFunction slideOverworld(boolean isAmplified, DensityFunction caves) {
-        return slide(caves, -64, 384, isAmplified ? 16 : 80, isAmplified ? 0 : 64, -0.078125F, 0, 24, isAmplified ? 0.4F : 0.1171875F);
+        return slide(caves, -64, 384, isAmplified ? 16 : 80, isAmplified ? 0 : 64, -0.078125, 0, 24, isAmplified ? 0.4 : 0.1171875);
     }
 
     private static DensityFunction slideNetherLike(HolderGetter<DensityFunction> functions, int minY, int height) {
-        return slide(getFunction(functions, BASE_3D_NOISE_NETHER), minY, height, 24, 0, 0.9375F, -8, 24, 2.5F);
+        return slide(getFunction(functions, BASE_3D_NOISE_NETHER), minY, height, 24, 0, 0.9375, -8, 24, 2.5);
     }
 
     private static DensityFunction slideEndLike(DensityFunction caves, int minY, int height) {
-        return slide(caves, minY, height, 72, -184, -23.4375F, 4, 32, -0.234375F);
+        return slide(caves, minY, height, 72, -184, -23.4375, 4, 32, -0.234375);
     }
 
-    protected static NoiseRouter nether(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise> noises) {
+    protected static NoiseRouter nether(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise.NoiseParameters> noises) {
         DensityFunction temperature = DensityFunctions.shiftedNoise2d(
             DensityFunctions.zero(), DensityFunctions.zero(), 0.25, noises.getOrThrow(Noises.TEMPERATURE_NETHER)
         );
@@ -475,7 +_,7 @@
             DensityFunctions.zero(), DensityFunctions.zero(), 0.25, noises.getOrThrow(Noises.VEGETATION_NETHER)
         );
         DensityFunction slide = slideNetherLike(functions, 0, 128);
-        DensityFunction fullNoise = DensityFunctions.add(postProcess(slide, 4, 8), DensityFunctions.beardifier());
+        DensityFunction fullNoise = DensityFunctions.add(postProcess(slide), DensityFunctions.beardifier());
         return new NoiseRouter(
             temperature,
             vegetation,
@@ -490,12 +_,12 @@
 
     protected static NoiseRouter caves(HolderGetter<DensityFunction> functions) {
         DensityFunction slide = slideNetherLike(functions, -64, 192);
-        return simpleRouter(DensityFunctions.add(postProcess(slide, 4, 8), DensityFunctions.beardifier()));
+        return simpleRouter(DensityFunctions.add(postProcess(slide), DensityFunctions.beardifier()));
     }
 
-    protected static NoiseRouter floatingIslands(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise> noises) {
+    protected static NoiseRouter floatingIslands(HolderGetter<DensityFunction> functions, HolderGetter<NormalNoise.NoiseParameters> noises) {
         DensityFunction slide = slideEndLike(getFunction(functions, BASE_3D_NOISE_END), 0, 256);
-        return simpleRouter(DensityFunctions.add(postProcess(slide, 8, 4), DensityFunctions.beardifier()));
+        return simpleRouter(DensityFunctions.add(postProcess(slide), DensityFunctions.beardifier()));
     }
 
     private static DensityFunction slideEnd(DensityFunction caves) {
@@ -503,8 +_,8 @@
     }
 
     protected static NoiseRouter end(HolderGetter<DensityFunction> functions) {
-        DensityFunction islands = getFunction(functions, END_ISLANDS);
-        DensityFunction fullNoise = DensityFunctions.add(postProcess(slideEnd(getFunction(functions, SLOPED_CHEESE_END)), 8, 4), DensityFunctions.beardifier());
+        DensityFunction islands = DensityFunctions.cache2d(DensityFunctions.endIslands(0L));
+        DensityFunction fullNoise = DensityFunctions.add(postProcess(slideEnd(getFunction(functions, SLOPED_CHEESE_END))), DensityFunctions.beardifier());
         return new NoiseRouter(
             DensityFunctions.zero(),
             DensityFunctions.zero(),
@@ -523,10 +_,10 @@
             DensityFunctions.zero(),
             DensityFunctions.zero(),
             DensityFunctions.zero(),
-            DensityFunctions.zero(),
-            DensityFunctions.zero(),
-            DensityFunctions.zero(),
-            fullNoise
+            fullNoise,
+            DensityFunctions.zero(),
+            DensityFunctions.zero(),
+            DensityFunctions.zero()
         );
     }
 
@@ -535,78 +_,81 @@
     }
 
     private static DensityFunction splineWithBlending(DensityFunction spline, DensityFunction blendingTarget) {
-        return DensityFunctions.cache(DensityFunctions.lerp(DensityFunctions.blendAlpha(), blendingTarget, spline));
+        DensityFunction blendedSpline = DensityFunctions.lerp(DensityFunctions.blendAlpha(), blendingTarget, spline);
+        return DensityFunctions.flatCache(DensityFunctions.cache2d(blendedSpline));
     }
 
     private static DensityFunction noiseGradientDensity(DensityFunction factor, DensityFunction depthWithJaggedness) {
         DensityFunction gradientUnscaled = DensityFunctions.mul(depthWithJaggedness, factor);
-        return gradientUnscaled.quarterNegative().mul(4.0F);
+        return gradientUnscaled.quarterNegative().mul(4.0);
     }
 
     private static DensityFunction preliminarySurfaceLevel(DensityFunction offset, DensityFunction factor, boolean amplified) {
+        DensityFunction cachedFactor = DensityFunctions.cache2d(factor);
+        DensityFunction cachedOffset = DensityFunctions.cache2d(offset);
         DensityFunction upperBound = DensityFunctions.remap(
-            DensityFunctions.sub(DensityFunctions.div(DensityFunctions.constant(0.2734375F), factor), offset), 1.5F, -1.5F, -64.0F, 320.0F
+            DensityFunctions.sub(DensityFunctions.div(DensityFunctions.constant(0.2734375), cachedFactor), cachedOffset), 1.5, -1.5, -64.0, 320.0
         );
-        upperBound = upperBound.clamp(-40.0F, 320.0F);
+        upperBound = upperBound.clamp(-40.0, 320.0);
         DensityFunction density = DensityFunctions.add(
             slideOverworld(
                 amplified,
-                DensityFunctions.add(noiseGradientDensity(factor, offsetToDepth(offset)), DensityFunctions.constant(-0.703125F)).clamp(-64.0F, 64.0F)
+                DensityFunctions.add(noiseGradientDensity(cachedFactor, offsetToDepth(cachedOffset)), DensityFunctions.constant(-0.703125)).clamp(-64.0, 64.0)
             ),
-            DensityFunctions.constant(-0.390625F)
+            DensityFunctions.constant(-0.390625)
         );
-        return DensityFunctions.findTopSurface(density, upperBound, -64, 8);
+        return DensityFunctions.findTopSurface(density, upperBound, -64, NoiseSettings.OVERWORLD_NOISE_SETTINGS.getCellHeight());
     }
 
     private static DensityFunction yLimitedInterpolatable(
-        DensityFunction y, DensityFunction whenInRange, int minYInclusive, int maxYInclusive, int whenOutOfRange, int cellSizeXz, int cellSizeY
+        DensityFunction y, DensityFunction whenInRange, int minYInclusive, int maxYInclusive, int whenOutOfRange
     ) {
         return DensityFunctions.interpolated(
-            DensityFunctions.rangeChoice(y, minYInclusive, maxYInclusive + 1, whenInRange, DensityFunctions.constant(whenOutOfRange)), cellSizeXz, cellSizeY
+            DensityFunctions.rangeChoice(y, minYInclusive, maxYInclusive + 1, whenInRange, DensityFunctions.constant(whenOutOfRange))
         );
     }
 
     private static DensityFunction slide(
-        DensityFunction caves, int minY, int height, int topStartY, int topEndY, float topTarget, int bottomStartY, int bottomEndY, float bottomTarget
+        DensityFunction caves, int minY, int height, int topStartY, int topEndY, double topTarget, int bottomStartY, int bottomEndY, double bottomTarget
     ) {
         DensityFunction noiseValue = caves;
-        DensityFunction topFactor = DensityFunctions.yClampedGradient(minY + height - topStartY, minY + height - topEndY, 1.0F, 0.0F);
+        DensityFunction topFactor = DensityFunctions.yClampedGradient(minY + height - topStartY, minY + height - topEndY, 1.0, 0.0);
         noiseValue = DensityFunctions.lerp(topFactor, topTarget, noiseValue);
-        DensityFunction bottomFactor = DensityFunctions.yClampedGradient(minY + bottomStartY, minY + bottomEndY, 0.0F, 1.0F);
+        DensityFunction bottomFactor = DensityFunctions.yClampedGradient(minY + bottomStartY, minY + bottomEndY, 0.0, 1.0);
         return DensityFunctions.lerp(bottomFactor, bottomTarget, noiseValue);
     }
 
     protected static final class QuantizedSpaghettiRarity {
-        public static DensityFunction wrapRarity2d(DensityFunction input, Holder<NormalNoise> noise) {
-            return DensityFunctions.intervalSelect(
-                    input,
-                    FloatList.of(-0.75F, -0.5F, 0.5F, 0.75F),
-                    List.of(
-                        noiseFunctionForRarity(noise, 0.5F),
-                        noiseFunctionForRarity(noise, 0.75F),
-                        noiseFunctionForRarity(noise, 1.0F),
-                        noiseFunctionForRarity(noise, 2.0F),
-                        noiseFunctionForRarity(noise, 3.0F)
-                    )
-                )
-                .abs();
-        }
-
-        public static DensityFunction wrapRarity3d(DensityFunction input, Holder<NormalNoise> noise) {
-            return DensityFunctions.intervalSelect(
-                    input,
-                    FloatList.of(-0.5F, 0.0F, 0.5F),
-                    List.of(
-                        noiseFunctionForRarity(noise, 0.75F),
-                        noiseFunctionForRarity(noise, 1.0F),
-                        noiseFunctionForRarity(noise, 1.5F),
-                        noiseFunctionForRarity(noise, 2.0F)
-                    )
-                )
-                .abs();
-        }
-
-        private static DensityFunction noiseFunctionForRarity(Holder<NormalNoise> noise, float rarity) {
+        public static DensityFunction wrapRarity2d(DensityFunction input, Holder<NormalNoise.NoiseParameters> noise) {
+            return DensityFunctions.intervalSelect(
+                    input,
+                    DoubleList.of(-0.75, -0.5, 0.5, 0.75),
+                    List.of(
+                        noiseFunctionForRarity(noise, 0.5),
+                        noiseFunctionForRarity(noise, 0.75),
+                        noiseFunctionForRarity(noise, 1.0),
+                        noiseFunctionForRarity(noise, 2.0),
+                        noiseFunctionForRarity(noise, 3.0)
+                    )
+                )
+                .abs();
+        }
+
+        public static DensityFunction wrapRarity3d(DensityFunction input, Holder<NormalNoise.NoiseParameters> noise) {
+            return DensityFunctions.intervalSelect(
+                    input,
+                    DoubleList.of(-0.5, 0.0, 0.5),
+                    List.of(
+                        noiseFunctionForRarity(noise, 0.75),
+                        noiseFunctionForRarity(noise, 1.0),
+                        noiseFunctionForRarity(noise, 1.5),
+                        noiseFunctionForRarity(noise, 2.0)
+                    )
+                )
+                .abs();
+        }
+
+        private static DensityFunction noiseFunctionForRarity(Holder<NormalNoise.NoiseParameters> noise, double rarity) {
             return DensityFunctions.noise(noise, 1.0 / rarity, 1.0 / rarity).mul(rarity);
         }
     }
