--- a/net/minecraft/world/level/levelgen/OverworldFunctionSet.java
+++ b/net/minecraft/world/level/levelgen/OverworldFunctionSet.java
@@ -3,18 +_,7 @@
 import java.util.function.Function;
 
 public record OverworldFunctionSet<T>(
-    T temperature,
-    T vegetation,
-    T continents,
-    T erosion,
-    T offset,
-    T factor,
-    T jaggedness,
-    T depth,
-    T slopedCheese,
-    T preliminarySurfaceLevel,
-    T chunkSurfaceLevel,
-    T finalDensity
+    T temperature, T vegetation, T continents, T erosion, T offset, T factor, T jaggedness, T depth, T slopedCheese, T preliminarySurfaceLevel
 ) {
     public <U> OverworldFunctionSet<U> map(Function<T, U> function) {
         return new OverworldFunctionSet<>(
@@ -27,9 +_,7 @@
             function.apply(this.jaggedness),
             function.apply(this.depth),
             function.apply(this.slopedCheese),
-            function.apply(this.preliminarySurfaceLevel),
-            function.apply(this.chunkSurfaceLevel),
-            function.apply(this.finalDensity)
+            function.apply(this.preliminarySurfaceLevel)
         );
     }
 }
