--- a/net/minecraft/world/level/levelgen/PatrolSpawner.java
+++ b/net/minecraft/world/level/levelgen/PatrolSpawner.java
@@ -83,6 +_,7 @@
 
             mob.setPos(pos.getX(), pos.getY(), pos.getZ());
             mob.finalizeSpawn(level, level.getCurrentDifficultyAt(pos), EntitySpawnReason.PATROL, null);
+            mob.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PATROL); // CraftBukkit
             level.addFreshEntityWithPassengers(mob);
             return true;
         } else {
