--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -5,6 +_,7 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.Arrays;
 import java.util.IdentityHashMap;
 import java.util.List;
 import java.util.Map;
@@ -578,7 +_,7 @@
         }
     }
 
-    public abstract static class LazyCondition implements SurfaceRules.Condition {
+    private abstract static class LazyCondition implements SurfaceRules.Condition {
         protected final SurfaceRules.Context context;
         private long lastUpdate;
         private @Nullable Boolean result;
@@ -815,7 +_,7 @@
         }
     }
 
-    public interface SurfaceRule {
+    protected interface SurfaceRule {
         @Nullable BlockState tryApply(final int blockX, final int blockY, final int blockZ);
     }
 
@@ -860,7 +_,7 @@
         }
     }
 
-    public record VerticalGradientConditionSource(Identifier randomName, VerticalAnchor trueAtAndBelow, VerticalAnchor falseAtAndAbove)
+    private record VerticalGradientConditionSource(Identifier randomName, VerticalAnchor trueAtAndBelow, VerticalAnchor falseAtAndAbove)
         implements SurfaceRules.ConditionSource {
         private static final MapCodec<SurfaceRules.VerticalGradientConditionSource> CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
