--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -244,7 +_,7 @@
         }
     }
 
-    public interface Condition {
+    private interface Condition {
         boolean test();
     }
 
@@ -270,7 +_,7 @@
         MapCodec<? extends SurfaceRules.ConditionSource> codec();
     }
 
-    public static final class Context {
+    protected static final class Context {
         private static final int HOW_FAR_BELOW_PRELIMINARY_SURFACE_LEVEL_TO_BUILD_SURFACE = 8;
         private static final int SURFACE_CELL_BITS = 4;
         private static final int SURFACE_CELL_SIZE = 16;
@@ -280,19 +_,19 @@
         private final SurfaceRules.Condition steep = new SurfaceRules.Context.SteepMaterialCondition(this);
         private final SurfaceRules.Condition hole = new SurfaceRules.Context.HoleCondition(this);
         private final SurfaceRules.Condition abovePreliminarySurface = new SurfaceRules.Context.AbovePreliminarySurfaceCondition();
-        public final RandomState randomState;
+        private final RandomState randomState;
         private final ChunkAccess chunk;
         private final NoiseChunk noiseChunk;
         private final Function<BlockPos, Holder<Biome>> biomeGetter;
-        public final WorldGenerationContext context;
+        private final WorldGenerationContext context;
         private final @Nullable Set<Holder<Biome>> possibleBiomes;
         private long lastPreliminarySurfaceCellOrigin = Long.MAX_VALUE;
         private final int[] preliminarySurfaceCache = new int[4];
         private final Map<ResourceKey<NormalNoise.NoiseParameters>, DoubleSupplier> noiseSamplers2d = new IdentityHashMap<>();
         private final Map<ResourceKey<NormalNoise.NoiseParameters>, DoubleSupplier> noiseSamplers3d = new IdentityHashMap<>();
         private long lastUpdateXZ = -9223372036854775807L;
-        public int blockX;
-        public int blockZ;
+        private int blockX;
+        private int blockZ;
         private int surfaceDepth;
         private long lastSurfaceDepth2Update = this.lastUpdateXZ - 1L;
         private double surfaceSecondary;
@@ -301,7 +_,7 @@
         private long lastUpdateY = -9223372036854775807L;
         private final BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
         private @Nullable Holder<Biome> biome;
-        public int blockY;
+        private int blockY;
         private int waterHeight;
         private int stoneDepthBelow;
         private int stoneDepthAbove;
@@ -528,7 +_,7 @@
         }
     }
 
-    public abstract static class LazyCondition implements SurfaceRules.Condition {
+    private abstract static class LazyCondition implements SurfaceRules.Condition {
         protected final SurfaceRules.Context context;
         private long lastUpdate;
         private @Nullable Boolean result;
@@ -570,7 +_,7 @@
         }
     }
 
-    public abstract static class LazyYCondition extends SurfaceRules.LazyCondition {
+    private abstract static class LazyYCondition extends SurfaceRules.LazyCondition {
         protected LazyYCondition(SurfaceRules.Context context) {
             super(context);
         }
@@ -750,7 +_,7 @@
         }
     }
 
-    public interface SurfaceRule {
+    protected interface SurfaceRule {
         @Nullable BlockState tryApply(final int blockX, final int blockY, final int blockZ);
     }
 
@@ -795,7 +_,7 @@
         }
     }
 
-    public record VerticalGradientConditionSource(Identifier randomName, VerticalAnchor trueAtAndBelow, VerticalAnchor falseAtAndAbove)
+    private record VerticalGradientConditionSource(Identifier randomName, VerticalAnchor trueAtAndBelow, VerticalAnchor falseAtAndAbove)
         implements SurfaceRules.ConditionSource {
         private static final MapCodec<SurfaceRules.VerticalGradientConditionSource> CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
