--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -528,7 +_,7 @@
         }
     }
 
-    public abstract static class LazyCondition implements SurfaceRules.Condition {
+    private abstract static class LazyCondition implements SurfaceRules.Condition {
         protected final SurfaceRules.Context context;
         private long lastUpdate;
         private @Nullable Boolean result;
@@ -750,7 +_,7 @@
         }
     }
 
-    public interface SurfaceRule {
+    protected interface SurfaceRule {
         @Nullable BlockState tryApply(final int blockX, final int blockY, final int blockZ);
     }
 
@@ -795,7 +_,7 @@
         }
     }
 
-    public record VerticalGradientConditionSource(Identifier randomName, VerticalAnchor trueAtAndBelow, VerticalAnchor falseAtAndAbove)
+    private record VerticalGradientConditionSource(Identifier randomName, VerticalAnchor trueAtAndBelow, VerticalAnchor falseAtAndAbove)
         implements SurfaceRules.ConditionSource {
         private static final MapCodec<SurfaceRules.VerticalGradientConditionSource> CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
