--- a/net/minecraft/client/data/models/model/ModelTemplate.java
+++ b/net/minecraft/client/data/models/model/ModelTemplate.java
@@ -20,9 +_,9 @@
 
 @OnlyIn(Dist.CLIENT)
 public class ModelTemplate {
-    public final Optional<Identifier> model;
-    public final Set<TextureSlot> requiredSlots;
-    public final Optional<String> suffix;
+    private final Optional<Identifier> model;
+    private final Set<TextureSlot> requiredSlots;
+    private final Optional<String> suffix;
 
     public ModelTemplate(Optional<Identifier> model, Optional<String> suffix, TextureSlot... requiredSlots) {
         this.model = model;
