--- a/net/minecraft/client/data/models/model/ModelTemplates.java
+++ b/net/minecraft/client/data/models/model/ModelTemplates.java
@@ -246,23 +_,23 @@
     public static final ModelTemplate FLAT_HANDHELD_MACE_ITEM = createItem("handheld_mace", TextureSlot.LAYER0);
     public static final ModelTemplate LIGHTNING_ROD = create("template_lightning_rod", TextureSlot.TEXTURE);
 
-    public static ModelTemplate create(TextureSlot... slots) {
+    private static ModelTemplate create(TextureSlot... slots) {
         return new ModelTemplate(Optional.empty(), Optional.empty(), slots);
     }
 
-    public static ModelTemplate create(String id, TextureSlot... slots) {
+    private static ModelTemplate create(String id, TextureSlot... slots) {
         return new ModelTemplate(Optional.of(Identifier.withDefaultNamespace("block/" + id)), Optional.empty(), slots);
     }
 
-    public static ModelTemplate createItem(String id, TextureSlot... slots) {
+    private static ModelTemplate createItem(String id, TextureSlot... slots) {
         return new ModelTemplate(Optional.of(Identifier.withDefaultNamespace("item/" + id)), Optional.empty(), slots);
     }
 
-    public static ModelTemplate createItem(String id, String suffix, TextureSlot... slots) {
+    private static ModelTemplate createItem(String id, String suffix, TextureSlot... slots) {
         return new ModelTemplate(Optional.of(Identifier.withDefaultNamespace("item/" + id)), Optional.of(suffix), slots);
     }
 
-    public static ModelTemplate create(String id, String suffix, TextureSlot... slots) {
+    private static ModelTemplate create(String id, String suffix, TextureSlot... slots) {
         return new ModelTemplate(Optional.of(Identifier.withDefaultNamespace("block/" + id)), Optional.of(suffix), slots);
     }
 }
