--- a/net/minecraft/client/data/models/model/TextureSlot.java
+++ b/net/minecraft/client/data/models/model/TextureSlot.java
@@ -54,11 +_,11 @@
     private final String id;
     private final @Nullable TextureSlot parent;
 
-    public static TextureSlot create(String id) {
+    private static TextureSlot create(String id) {
         return new TextureSlot(id, null);
     }
 
-    public static TextureSlot create(String id, TextureSlot parent) {
+    private static TextureSlot create(String id, TextureSlot parent) {
         return new TextureSlot(id, parent);
     }
 
