--- a/net/minecraft/client/data/models/model/TexturedModel.java
+++ b/net/minecraft/client/data/models/model/TexturedModel.java
@@ -9,11 +_,8 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
-import net.neoforged.neoforge.client.extensions.ITexturedModelExtension;
-import net.neoforged.neoforge.client.extensions.ITexturedModelExtension.ProviderExtension;
-
 @OnlyIn(Dist.CLIENT)
-public class TexturedModel implements ITexturedModelExtension {
+public class TexturedModel {
     public static final TexturedModel.Provider CUBE = createDefault(TextureMapping::cube, ModelTemplates.CUBE_ALL);
     public static final TexturedModel.Provider CUBE_INNER_FACES = createDefault(TextureMapping::cube, ModelTemplates.CUBE_ALL_INNER_FACES);
     public static final TexturedModel.Provider CUBE_MIRRORED = createDefault(TextureMapping::cube, ModelTemplates.CUBE_MIRRORED_ALL);
@@ -49,7 +_,7 @@
     private final TextureMapping mapping;
     private final ModelTemplate template;
 
-    public TexturedModel(TextureMapping mapping, ModelTemplate template) {
+    private TexturedModel(TextureMapping mapping, ModelTemplate template) {
         this.mapping = mapping;
         this.template = template;
     }
@@ -75,7 +_,7 @@
         return this.template.createWithSuffix(block, extraSuffix, this.mapping, modelOutput);
     }
 
-    public static TexturedModel.Provider createDefault(Function<Block, TextureMapping> mapping, ModelTemplate template) {
+    private static TexturedModel.Provider createDefault(Function<Block, TextureMapping> mapping, ModelTemplate template) {
         return block -> new TexturedModel(mapping.apply(block), template);
     }
 
@@ -84,7 +_,7 @@
     }
 
     @FunctionalInterface
-    public interface Provider extends ProviderExtension {
+    public interface Provider {
         TexturedModel get(final Block block);
 
         default Identifier create(Block block, BiConsumer<Identifier, ModelInstance> modelOutput) {
