--- a/com/mojang/renderpearl/api/commands/CommandEncoder.java
+++ b/com/mojang/renderpearl/api/commands/CommandEncoder.java
@@ -57,15 +_,7 @@
     void clearColorAndDepthTextures(GpuTexture colorTexture, Vector4fc clearColor, GpuTexture depthTexture, double clearDepth);
 
     void clearColorAndDepthTextures(
-        GpuTexture colorTexture,
-        Vector4fc clearColor,
-        GpuTexture depthTexture,
-        double clearDepth,
-        int regionX,
-        int regionY,
-        int regionWidth,
-        int regionHeight,
-        int mipLevel
+        GpuTexture colorTexture, Vector4fc clearColor, GpuTexture depthTexture, double clearDepth, int regionX, int regionY, int regionWidth, int regionHeight
     );
 
     void clearDepthTexture(GpuTexture depthTexture, double clearDepth);
