--- a/com/mojang/renderpearl/api/commands/GpuQuery.java
+++ b/com/mojang/renderpearl/api/commands/GpuQuery.java
@@ -1,12 +_,11 @@
 package com.mojang.renderpearl.api.commands;
 
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.util.OptionalLong;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface GpuQuery extends UncheckedAutoCloseable {
+public interface GpuQuery extends AutoCloseable {
     OptionalLong getValue();
 
     @Override
