--- a/com/mojang/renderpearl/api/commands/RenderPassDescriptor.java
+++ b/com/mojang/renderpearl/api/commands/RenderPassDescriptor.java
@@ -1,8 +_,6 @@
 package com.mojang.renderpearl.api.commands;
 
 import com.mojang.renderpearl.api.textures.GpuTextureView;
-import it.unimi.dsi.fastutil.objects.ReferenceArrayList;
-import it.unimi.dsi.fastutil.objects.ReferenceLists;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
@@ -21,7 +_,7 @@
     RenderPass.RenderArea renderArea
 ) {
     public RenderPassDescriptor {
-        colorAttachments = ReferenceLists.unmodifiable(new ReferenceArrayList<>(colorAttachments));
+        colorAttachments = List.copyOf(colorAttachments);
     }
 
     public static RenderPassDescriptor.Builder builder(Supplier<String> label) {
