--- a/net/minecraft/server/players/CachedUserNameToIdResolver.java
+++ b/net/minecraft/server/players/CachedUserNameToIdResolver.java
@@ -155,7 +_,7 @@
                 JsonArray entryList = this.gson.fromJson(reader, JsonArray.class);
                 if (entryList != null) {
                     DateFormat dateFormat = createDateFormat();
-                    entryList.forEach(element -> readGameProfile(element, dateFormat).ifPresent(result::add));
+                    entryList.forEach(element -> CachedUserNameToIdResolver.<CachedUserNameToIdResolver.GameProfileInfo>readGameProfile(element, dateFormat).ifPresent(result::add)); // CraftBukkit - decompile error
                     return result;
                 }
 
