--- a/net/minecraft/server/players/GameProfileCache.java
+++ b/net/minecraft/server/players/GameProfileCache.java
@@ -57,6 +_,11 @@
     @Nullable
     private Executor executor;
 
+    // Paper start - Fix GameProfileCache concurrency
+    protected final java.util.concurrent.locks.ReentrantLock stateLock = new java.util.concurrent.locks.ReentrantLock();
+    protected final java.util.concurrent.locks.ReentrantLock lookupLock = new java.util.concurrent.locks.ReentrantLock();
+    // Paper end - Fix GameProfileCache concurrency
+
     public GameProfileCache(GameProfileRepository p_10974_, File p_10975_) {
         this.profileRepository = p_10974_;
         this.file = p_10975_;
@@ -86,6 +_,8 @@
                     atomicreference.set(null);
                 }
             };
+            if (!org.apache.commons.lang3.StringUtils.isBlank(p_10995_) // Paper - Don't lookup a profile with a blank name
+                    && io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode()) // Paper - Add setting for proxy online mode status
             p_10994_.findProfilesByNames(new String[]{p_10995_}, profilelookupcallback);
             GameProfile gameprofile = atomicreference.get();
             return gameprofile != null ? Optional.of(gameprofile) : createUnknownProfile(p_10995_);
@@ -101,7 +_,7 @@
     }
 
     private static boolean usesAuthentication() {
-        return usesAuthentication;
+        return io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode(); // Paper - Add setting for proxy online mode status
     }
 
     public void add(GameProfile p_10992_) {
@@ -111,13 +_,26 @@
         Date date = calendar.getTime();
         GameProfileCache.GameProfileInfo gameprofilecache$gameprofileinfo = new GameProfileCache.GameProfileInfo(p_10992_, date);
         this.safeAdd(gameprofilecache$gameprofileinfo);
-        this.save();
+        if( !org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly ) this.save(); // Spigot - skip saving if disabled
     }
 
     private long getNextOperation() {
         return this.operationCount.incrementAndGet();
     }
 
+    // Paper start
+    public @Nullable GameProfile getProfileIfCached(String name) {
+        try { this.stateLock.lock(); // Paper - Fix GameProfileCache concurrency
+            GameProfileCache.GameProfileInfo entry = this.profilesByName.get(name.toLowerCase(Locale.ROOT));
+            if (entry == null) {
+                return null;
+            }
+            entry.setLastAccess(this.getNextOperation());
+            return entry.getProfile();
+        } finally { this.stateLock.unlock(); } // Paper - Fix GameProfileCache concurrency
+    }
+    // Paper end
+
     public Optional<GameProfile> get(String p_10997_) {
         String s = p_10997_.toLowerCase(Locale.ROOT);
         GameProfileCache.GameProfileInfo gameprofilecache$gameprofileinfo = this.profilesByName.get(s);
@@ -141,7 +_,7 @@
             }
         }
 
-        if (flag) {
+        if (flag && !org.spigotmc.SpigotConfig.saveUserCacheOnStopOnly) { // Spigot - skip saving if disabled
             this.save();
         }
 
@@ -157,7 +_,7 @@
                 return completablefuture;
             } else {
                 CompletableFuture<Optional<GameProfile>> completablefuture1 = CompletableFuture.<Optional<GameProfile>>supplyAsync(
-                        () -> this.get(p_143968_), Util.backgroundExecutor()
+                        () -> this.get(p_143968_), Util.PROFILE_EXECUTOR
                     )
                     .whenCompleteAsync((p_143965_, p_143966_) -> this.requests.remove(p_143968_), this.executor);
                 this.requests.put(p_143968_, completablefuture1);
@@ -216,7 +_,7 @@
     public void save() {
         JsonArray jsonarray = new JsonArray();
         DateFormat dateformat = createDateFormat();
-        this.getTopMRUProfiles(1000).forEach(p_143962_ -> jsonarray.add(writeGameProfile(p_143962_, dateformat)));
+        this.getTopMRUProfiles(org.spigotmc.SpigotConfig.userCacheCap).forEach(p_143962_ -> jsonarray.add(writeGameProfile(p_143962_, dateformat))); // Spigot
         String s = this.gson.toJson((JsonElement)jsonarray);
 
         try (Writer writer = Files.newWriter(this.file, StandardCharsets.UTF_8)) {
