--- a/net/minecraft/server/players/OldUsersConverter.java
+++ b/net/minecraft/server/players/OldUsersConverter.java
@@ -49,7 +_,7 @@
 
     private static void lookupPlayers(MinecraftServer p_11087_, Collection<String> p_11088_, ProfileLookupCallback p_11089_) {
         String[] astring = p_11088_.stream().filter(p_11077_ -> !StringUtil.isNullOrEmpty(p_11077_)).toArray(String[]::new);
-        if (p_11087_.usesAuthentication()) {
+        if (p_11087_.usesAuthentication() || org.spigotmc.SpigotConfig.bungee) { // Spigot: bungee = online mode, for now.
             p_11087_.getProfileRepository().findProfilesByNames(astring, p_11089_);
         } else {
             for (String s : astring) {
