--- a/net/minecraft/server/players/StoredUserList.java
+++ b/net/minecraft/server/players/StoredUserList.java
@@ -153,6 +_,14 @@
                         this.map.put(this.getKeyForUser(entry.getUser()), (V)entry);
                     }
                 }
+                // Spigot Start
+            } catch ( com.google.gson.JsonParseException | NullPointerException ex )
+            {
+                org.bukkit.Bukkit.getLogger().log( java.util.logging.Level.WARNING, "Unable to read file " + this.file + ", backing it up to {0}.backup and creating new copy.", ex );
+                File backup = new File( this.file + ".backup" );
+                this.file.renameTo( backup );
+                this.file.delete();
+                // Spigot End
             }
         }
     }
