--- a/net/minecraft/advancements/predicates/ItemPredicate.java
+++ b/net/minecraft/advancements/predicates/ItemPredicate.java
@@ -6,8 +_,8 @@
 import java.util.function.Predicate;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.tags.TagKey;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemInstance;
@@ -16,7 +_,7 @@
 public record ItemPredicate(Optional<HolderSet<Item>> items, MinMaxBounds.Ints count, DataComponentMatchers components) implements Predicate<ItemInstance> {
     public static final Codec<ItemPredicate> CODEC = RecordCodecBuilder.create(
         i -> i.group(
-                RegistryCodecs.holderSet(Registries.ITEM).optionalFieldOf("items").forGetter(ItemPredicate::items),
+                RegistryCodecs.homogeneousList(Registries.ITEM).optionalFieldOf("items").forGetter(ItemPredicate::items),
                 MinMaxBounds.Ints.CODEC.optionalFieldOf("count", MinMaxBounds.Ints.ANY).forGetter(ItemPredicate::count),
                 DataComponentMatchers.CODEC.forGetter(ItemPredicate::components)
             )
