--- a/net/minecraft/advancements/predicates/TagPredicate.java
+++ b/net/minecraft/advancements/predicates/TagPredicate.java
@@ -6,7 +_,7 @@
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderSet;
 import net.minecraft.core.Registry;
-import net.minecraft.core.registries.codec.RegistryCodecs;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.tags.TagKey;
 
@@ -14,7 +_,7 @@
     public static <T> Codec<TagPredicate<T>> codec(ResourceKey<? extends Registry<T>> registryKey) {
         return RecordCodecBuilder.create(
             i -> i.group(
-                    RegistryCodecs.holderSet(registryKey).fieldOf("id").forGetter(TagPredicate::tag),
+                    RegistryCodecs.homogeneousList(registryKey).fieldOf("id").forGetter(TagPredicate::tag),
                     Codec.BOOL.fieldOf("expected").forGetter(TagPredicate::expected)
                 )
                 .apply(i, TagPredicate::new)
