--- a/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
+++ b/net/minecraft/world/level/chunk/status/ChunkStatusTasks.java
@@ -42,7 +_,7 @@
         WorldGenContext context, ChunkStep step, StaticCache2D<GenerationChunkHolder> chunks, ChunkAccess chunk
     ) {
         ServerLevel level = context.level();
-        if (level.getServer().getWorldGenSettings().options().generateStructures()) {
+        if (level.worldGenSettings.options().generateStructures()) { // CraftBukkit
             context.generator()
                 .createStructures(
                     level.registryAccess(),
@@ -187,25 +_,80 @@
             } else {
                 levelChunk = new LevelChunk(level, protoChunk, lc -> {
                     try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(chunk.problemPath(), LOGGER)) {
-                        postLoadProtoChunk(level, TagValueInput.create(reporter, level.registryAccess(), protoChunk.getEntities()));
+                        postLoadProtoChunk(level, TagValueInput.create(reporter, level.registryAccess(), protoChunk.getEntities()), protoChunk.getPos()); // Paper - rewrite chunk system
                     }
                 });
                 holder.replaceProtoChunk(new ImposterProtoChunk(levelChunk, false));
             }
 
             levelChunk.setFullStatus(holder::getFullStatus);
-            levelChunk.runPostLoad();
+            try {
+                holder.currentlyLoading = levelChunk; // Neo: bypass the future chain when getChunk is called, this prevents deadlocks.
+                levelChunk.runPostLoad();
+            } finally {
+                holder.currentlyLoading = null; // Neo: Stop bypassing the future chain.
+            }
             levelChunk.setLoaded(true);
-            levelChunk.registerAllBlockEntitiesAfterLevelLoad();
-            levelChunk.registerTickContainerInLevel(level);
-            levelChunk.setUnsavedListener(context.unsavedListener());
+            try {
+                holder.currentlyLoading = levelChunk; // Neo: bypass the future chain when getChunk is called, this prevents deadlocks.
+                levelChunk.registerAllBlockEntitiesAfterLevelLoad();
+                levelChunk.registerTickContainerInLevel(level);
+                levelChunk.setUnsavedListener(context.unsavedListener());
+                net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.ChunkEvent.Load(levelChunk, !(protoChunk instanceof ImposterProtoChunk)));
+            } finally {
+                holder.currentlyLoading = null; // Neo: Stop bypassing the future chain.
+            }
             return levelChunk;
         }, context.mainThreadExecutor());
     }
 
-    public static void postLoadProtoChunk(ServerLevel level, ValueInput.ValueInputList entities) {
+    public static void postLoadProtoChunk(ServerLevel level, ValueInput.ValueInputList entities, final ChunkPos pos) { // Paper - rewrite chunk system - add ChunkPos param
         if (!entities.isEmpty()) {
-            level.addWorldGenChunkEntities(EntityType.loadEntitiesRecursive(entities, level, EntitySpawnReason.LOAD));
-        }
-    }
+            // Paper start - duplicate uuid resolving
+            level.addWorldGenChunkEntities(EntityType.loadEntitiesRecursive(entities, level, EntitySpawnReason.LOAD).filter((entity) -> {
+                return !checkDupeUUID(level, entity);
+            }), pos); // Paper - rewrite chunk system
+            // Paper end - duplicate uuid resolving
+        }
+    }
+
+    // Paper start - duplicate uuid resolving
+    // rets true if to prevent the entity from being added
+    public static boolean checkDupeUUID(ServerLevel level, net.minecraft.world.entity.Entity entity) {
+        io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DuplicateUUID.DuplicateUUIDMode mode = level.paperConfig().entities.spawning.duplicateUuid.mode;
+        if (mode != io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DuplicateUUID.DuplicateUUIDMode.WARN
+                && mode != io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DuplicateUUID.DuplicateUUIDMode.DELETE
+                && mode != io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DuplicateUUID.DuplicateUUIDMode.SAFE_REGEN) {
+            return false;
+        }
+        net.minecraft.world.entity.Entity other = level.getEntity(entity.getUUID());
+
+        if (other == null || other == entity) {
+            return false;
+        }
+
+        if (mode == io.papermc.paper.configuration.WorldConfiguration.Entities.Spawning.DuplicateUUID.DuplicateUUIDMode.SAFE_REGEN && other != null && !other.isRemoved()
+                && java.util.Objects.equals(other.getEncodeId(), entity.getEncodeId())
+                && entity.getBukkitEntity().getLocation().distance(other.getBukkitEntity().getLocation()) < level.paperConfig().entities.spawning.duplicateUuid.safeRegenDeleteRange
+        ) {
+            entity.discard(null);
+            return true;
+        }
+        if (!other.isRemoved()) {
+            switch (mode) {
+                case SAFE_REGEN: {
+                    entity.setUUID(java.util.UUID.randomUUID());
+                    break;
+                }
+                case DELETE: {
+                    entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
+                    return true;
+                }
+                default:
+                    break;
+            }
+        }
+        return false;
+    }
+    // Paper end - duplicate uuid resolving
 }
