--- a/net/minecraft/world/level/chunk/status/ChunkPyramid.java
+++ b/net/minecraft/world/level/chunk/status/ChunkPyramid.java
@@ -50,7 +_,7 @@
         .step(ChunkStatus.CARVERS, s -> s)
         .step(ChunkStatus.FEATURES, s -> s)
         .step(ChunkStatus.INITIALIZE_LIGHT, s -> s.setTask(ChunkStatusTasks::initializeLight))
-        .step(ChunkStatus.LIGHT, s -> s.addRequirement(ChunkStatus.INITIALIZE_LIGHT, 1).setTask(ChunkStatusTasks::light))
+        .step(ChunkStatus.LIGHT, s -> s.setTask(ChunkStatusTasks::light)) // Paper - rewrite chunk system - starlight does not need neighbours
         .step(ChunkStatus.SPAWN, s -> s)
         .step(ChunkStatus.FULL, s -> s.setTask(ChunkStatusTasks::full))
         .build();
