--- a/net/minecraft/world/level/chunk/status/ChunkPyramid.java
+++ b/net/minecraft/world/level/chunk/status/ChunkPyramid.java
@@ -14,16 +_,24 @@
         .step(ChunkStatus.STRUCTURE_REFERENCES, s -> s.addRequirement(ChunkStatus.STRUCTURE_STARTS, 8).setTask(ChunkStatusTasks::generateStructureReferences))
         .step(ChunkStatus.BIOMES, s -> s.addRequirement(ChunkStatus.STRUCTURE_STARTS, 8).setTask(ChunkStatusTasks::generateBiomes))
         .step(
-            ChunkStatus.TERRAIN,
-            s -> s.addRequirement(ChunkStatus.STRUCTURE_STARTS, 8)
-                .addRequirement(ChunkStatus.BIOMES, 1)
-                .blockStateWriteRadius(0)
-                .setTask(ChunkStatusTasks::buildTerrain)
-        )
+            ChunkStatus.NOISE,
+            s -> s.addRequirement(ChunkStatus.STRUCTURE_STARTS, 8)
+                .addRequirement(ChunkStatus.BIOMES, 1)
+                .blockStateWriteRadius(0)
+                .setTask(ChunkStatusTasks::generateNoise)
+        )
+        .step(
+            ChunkStatus.SURFACE,
+            s -> s.addRequirement(ChunkStatus.STRUCTURE_STARTS, 8)
+                .addRequirement(ChunkStatus.BIOMES, 1)
+                .blockStateWriteRadius(0)
+                .setTask(ChunkStatusTasks::generateSurface)
+        )
+        .step(ChunkStatus.CARVERS, s -> s.addRequirement(ChunkStatus.STRUCTURE_STARTS, 8).blockStateWriteRadius(0).setTask(ChunkStatusTasks::generateCarvers))
         .step(
             ChunkStatus.FEATURES,
             s -> s.addRequirement(ChunkStatus.STRUCTURE_STARTS, 8)
-                .addRequirement(ChunkStatus.TERRAIN, 1)
+                .addRequirement(ChunkStatus.CARVERS, 1)
                 .blockStateWriteRadius(1)
                 .setTask(ChunkStatusTasks::generateFeatures)
         )
@@ -37,10 +_,12 @@
         .step(ChunkStatus.STRUCTURE_STARTS, s -> s.setTask(ChunkStatusTasks::loadStructureStarts))
         .step(ChunkStatus.STRUCTURE_REFERENCES, s -> s)
         .step(ChunkStatus.BIOMES, s -> s)
-        .step(ChunkStatus.TERRAIN, s -> s)
+        .step(ChunkStatus.NOISE, s -> s)
+        .step(ChunkStatus.SURFACE, s -> s)
+        .step(ChunkStatus.CARVERS, s -> s)
         .step(ChunkStatus.FEATURES, s -> s)
         .step(ChunkStatus.INITIALIZE_LIGHT, s -> s.setTask(ChunkStatusTasks::initializeLight))
-        .step(ChunkStatus.LIGHT, s -> s.addRequirement(ChunkStatus.INITIALIZE_LIGHT, 1).setTask(ChunkStatusTasks::light))
+        .step(ChunkStatus.LIGHT, s -> s.setTask(ChunkStatusTasks::light)) // Paper - rewrite chunk system - starlight does not need neighbours
         .step(ChunkStatus.SPAWN, s -> s)
         .step(ChunkStatus.FULL, s -> s.setTask(ChunkStatusTasks::full))
         .build();
