--- a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+++ b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
@@ -115,7 +_,7 @@
             List<Entity> list = p_155911_.getEntities(null, PistonMath.getMovementArea(aabb, direction, d0).minmax(aabb));
             if (!list.isEmpty()) {
                 List<AABB> list1 = voxelshape.toAabbs();
-                boolean flag = p_155914_.movedState.is(Blocks.SLIME_BLOCK);
+                boolean flag = p_155914_.movedState.isSlimeBlock(); //TODO: is this patch really needed the logic of the original seems sound revisit later
                 Iterator iterator = list.iterator();
 
                 while (true) {
@@ -148,6 +_,10 @@
                                 }
 
                                 entity.setDeltaMovement(d1, d2, d3);
+                                // Paper - EAR items stuck in in slime pushed by a piston
+                                entity.activatedTick = Math.max(entity.activatedTick, net.minecraft.server.MinecraftServer.currentTick + 10);
+                                entity.activatedImmunityTick = Math.max(entity.activatedImmunityTick, net.minecraft.server.MinecraftServer.currentTick + 10);
+                                // Paper end
                                 break;
                             }
                         }
@@ -298,7 +_,7 @@
                 if (p_155916_.getBlockState(p_155917_).is(Blocks.MOVING_PISTON)) {
                     BlockState blockstate = Block.updateFromNeighbourShapes(p_155919_.movedState, p_155916_, p_155917_);
                     if (blockstate.isAir()) {
-                        p_155916_.setBlock(p_155917_, p_155919_.movedState, 84);
+                        p_155916_.setBlock(p_155917_, p_155919_.movedState, io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPistonDuplication ? 84 : (84 | Block.UPDATE_CLIENTS)); // Paper - fix a variety of piston desync dupes; force notify (flag 2), it's possible the set type by the piston block (which doesn't notify) set this block to air
                         Block.updateOrDestroy(p_155919_.movedState, blockstate, p_155916_, p_155917_, 3);
                     } else {
                         if (blockstate.hasProperty(BlockStateProperties.WATERLOGGED) && blockstate.getValue(BlockStateProperties.WATERLOGGED)) {
