--- a/net/minecraft/world/entity/ai/goal/target/PathfinderGoalDefendVillage.java
+++ b/net/minecraft/world/entity/ai/goal/target/PathfinderGoalDefendVillage.java
@@ -30,7 +30,7 @@
     public boolean canUse() {
         AxisAlignedBB axisalignedbb = this.golem.getBoundingBox().inflate(10.0D, 8.0D, 10.0D);
         WorldServer worldserver = getServerLevel((Entity) this.golem);
-        List<? extends EntityLiving> list = worldserver.<EntityLiving>getNearbyEntities(EntityVillager.class, this.attackTargeting, this.golem, axisalignedbb);
+        List<? extends EntityLiving> list = worldserver.getNearbyEntities(EntityVillager.class, this.attackTargeting, this.golem, axisalignedbb); // CraftBukkit - decompile error
         List<EntityHuman> list1 = worldserver.getNearbyPlayers(this.attackTargeting, this.golem, axisalignedbb);
 
         for (EntityLiving entityliving : list) {
@@ -64,7 +64,7 @@
 
     @Override
     public void start() {
-        this.golem.setTarget(this.potentialTarget);
+        this.golem.setTarget(this.potentialTarget, org.bukkit.event.entity.EntityTargetEvent.TargetReason.DEFEND_VILLAGE, true); // CraftBukkit - reason
         super.start();
     }
 }
