--- a/net/minecraft/client/gui/screens/inventory/AbstractSignEditScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/AbstractSignEditScreen.java
@@ -55,7 +_,6 @@
         this.slot = slot;
         this.woodType = SignBlock.getWoodType(sign.getBlockState().getBlock());
         this.textColor = currentText.hasGlowingText() ? currentText.getColor().getTextColor() : AbstractSignRenderer.getDarkColor(currentText);
-        this.text.setColor(currentText.getColor()).setTextGlowing(currentText.hasGlowingText());
 
         for (int i = 0; i < 4; i++) {
             String stringifiedLine = i < currentLines.size() ? currentLines.get(i).getString() : "";
