--- a/net/minecraft/client/gui/screens/inventory/CartographyTableScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/CartographyTableScreen.java
@@ -6,7 +_,6 @@
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.resources.Identifier;
-import net.minecraft.tags.ItemTags;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.inventory.CartographyTableMenu;
 import net.minecraft.world.item.ItemStack;
@@ -57,7 +_,7 @@
                     }
                 }
 
-                if (isScaling && (!map.is(ItemTags.EXTENDABLE_MAPS) || mapData.scale >= 4)) {
+                if (isScaling && mapData.scale >= 4) {
                     locked = true;
                     graphics.blitSprite(RenderPipelines.GUI_TEXTURED, ERROR_SPRITE, xo + 35, yo + 31, 28, 21);
                 }
