--- a/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
@@ -37,7 +_,6 @@
 import net.minecraft.resources.Identifier;
 import net.minecraft.tags.TagKey;
 import net.minecraft.util.Mth;
-import net.minecraft.util.Prediction;
 import net.minecraft.util.Unit;
 import net.minecraft.world.Container;
 import net.minecraft.world.SimpleContainer;
@@ -205,14 +_,14 @@
                     }
 
                     if (buttonNum == 0) {
-                        this.minecraft.player.drop(this.menu.getCarried(), true, Prediction.PREDICTED);
+                        this.minecraft.player.drop(this.menu.getCarried(), true);
                         this.minecraft.gameMode.handleCreativeModeItemDrop(this.menu.getCarried());
                         this.menu.setCarried(ItemStack.EMPTY);
                     }
 
                     if (buttonNum == 1) {
                         ItemStack removedItem = this.menu.getCarried().split(1);
-                        this.minecraft.player.drop(removedItem, true, Prediction.PREDICTED);
+                        this.minecraft.player.drop(removedItem, true);
                         this.minecraft.gameMode.handleCreativeModeItemDrop(removedItem);
                     }
                 }
@@ -232,11 +_,11 @@
                     } else if (containerInput == ContainerInput.THROW && slot != null && slot.hasItem()) {
                         ItemStack toDrop = slot.remove(buttonNum == 0 ? 1 : slot.getItem().getMaxStackSize());
                         ItemStack afterDrop = slot.getItem();
-                        this.minecraft.player.drop(toDrop, true, Prediction.PREDICTED);
+                        this.minecraft.player.drop(toDrop, true);
                         this.minecraft.gameMode.handleCreativeModeItemDrop(toDrop);
                         this.minecraft.gameMode.handleCreativeModeItemAdd(afterDrop, ((CreativeModeInventoryScreen.SlotWrapper)slot).target.index);
                     } else if (containerInput == ContainerInput.THROW && slotId == -999 && !this.menu.getCarried().isEmpty()) {
-                        this.minecraft.player.drop(this.menu.getCarried(), true, Prediction.PREDICTED);
+                        this.minecraft.player.drop(this.menu.getCarried(), true);
                         this.minecraft.gameMode.handleCreativeModeItemDrop(this.menu.getCarried());
                         this.menu.setCarried(ItemStack.EMPTY);
                     } else {
@@ -275,7 +_,7 @@
                     if (containerInput == ContainerInput.THROW) {
                         if (!clicked.isEmpty()) {
                             ItemStack toDrop = clicked.copyWithCount(buttonNum == 0 ? 1 : clicked.getMaxStackSize());
-                            this.minecraft.player.drop(toDrop, true, Prediction.PREDICTED);
+                            this.minecraft.player.drop(toDrop, true);
                             this.minecraft.gameMode.handleCreativeModeItemDrop(toDrop);
                         }
 
@@ -312,7 +_,7 @@
                             int numToDrop = buttonNum == 0 ? 1 : oldItemStack.getCount();
                             ItemStack toDrop = oldItemStack.copyWithCount(numToDrop);
                             oldItemStack.shrink(numToDrop);
-                            this.minecraft.player.drop(toDrop, true, Prediction.PREDICTED);
+                            this.minecraft.player.drop(toDrop, true);
                             this.minecraft.gameMode.handleCreativeModeItemDrop(toDrop);
                         }
 
