--- a/net/minecraft/client/gui/screens/inventory/SmithingScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/SmithingScreen.java
@@ -134,7 +_,7 @@
                     if (HumanoidArmorLayer.shouldRender(itemStack, EquipmentSlot.HEAD)) {
                         this.armorStandPreview.headEquipment = itemStack.copy();
                     } else {
-                        itemModelResolver.updateForTopItem(this.armorStandPreview.headItem, itemStack, ItemDisplayContext.HEAD, this.minecraft.level, null, 0);
+                        itemModelResolver.updateForTopItem(this.armorStandPreview.headItem, itemStack, ItemDisplayContext.HEAD, null, null, 0);
                     }
                     break;
                 case CHEST:
@@ -150,7 +_,7 @@
                 default:
                     this.armorStandPreview.leftHandItemStack = itemStack.copy();
                     itemModelResolver.updateForTopItem(
-                        this.armorStandPreview.leftHandItemState, itemStack, ItemDisplayContext.THIRD_PERSON_LEFT_HAND, this.minecraft.level, null, 0
+                        this.armorStandPreview.leftHandItemState, itemStack, ItemDisplayContext.THIRD_PERSON_LEFT_HAND, null, null, 0
                     );
             }
         }
