--- a/net/minecraft/world/entity/animal/cow/MushroomCow.java
+++ b/net/minecraft/world/entity/animal/cow/MushroomCow.java
@@ -13,7 +_,6 @@
 import net.minecraft.core.particles.SpellParticleOption;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
-import net.minecraft.network.protocol.game.ClientboundLevelParticlesPacket;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
@@ -54,6 +_,7 @@
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.jspecify.annotations.Nullable;
 
 public class MushroomCow extends AbstractCow implements Shearable {
@@ -70,13 +_,58 @@
         super(type, level);
     }
 
+    // Purpur start - Ridables
+    @Override
+    public boolean isRidable() {
+        return level().purpurConfig.mooshroomRidable;
+    }
+
+    @Override
+    public boolean dismountsUnderwater() {
+        return level().purpurConfig.useDismountsUnderwaterTag ? super.dismountsUnderwater() : !level().purpurConfig.mooshroomRidableInWater;
+    }
+
+    @Override
+    public boolean isControllable() {
+        return level().purpurConfig.mooshroomControllable;
+    }
+    // Purpur end - Ridables
+
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(net.minecraft.world.entity.ai.attributes.Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.mooshroomMaxHealth);
+    }
+    // Purpur end - Configurable entity base attributes
+
+    // Purpur start - Make entity breeding times configurable
+    @Override
+    public int getPurpurBreedTime() {
+        return this.level().purpurConfig.mooshroomBreedingTicks;
+    }
+    // Purpur end - Make entity breeding times configurable
+
+    // Purpur start - Toggle for water sensitive mob damage
+    @Override
+    public boolean isSensitiveToWater() {
+        return this.level().purpurConfig.mooshroomTakeDamageFromWater;
+    }
+    // Purpur end - Toggle for water sensitive mob damage
+
+    // Purpur start - Mobs always drop experience
+    @Override
+    protected boolean isAlwaysExperienceDropper() {
+        return this.level().purpurConfig.mooshroomAlwaysDropExp;
+    }
+    // Purpur end - Mobs always drop experience
+
     @Override
     public float getWalkTargetValue(BlockPos pos, LevelReader level) {
         return level.getBlockState(pos.below()).is(Blocks.MYCELIUM) ? 10.0F : level.getPathfindingCostFromLightLevels(pos);
     }
 
     public static boolean checkMushroomSpawnRules(
-        EntityType<MushroomCow> type, LevelAccessor level, EntitySpawnReason spawnReason, BlockPos pos, RandomSource random
+        EntityType<? extends MushroomCow> type, LevelAccessor level, EntitySpawnReason spawnReason, BlockPos pos, RandomSource random
     ) {
         return level.getBlockState(pos.below()).is(BlockTags.MOOSHROOMS_SPAWNABLE_ON) && isBrightEnoughToSpawn(level, pos);
     }
@@ -125,7 +_,13 @@
             return InteractionResult.SUCCESS;
         } else if (itemStack.is(Items.SHEARS) && this.readyForShearing()) {
             if (this.level() instanceof ServerLevel level) {
-                this.shear(level, SoundSource.PLAYERS, itemStack);
+                // Paper start - call PlayerShearEntityEvent
+                java.util.List<ItemStack> drops = this.generateDefaultDrops(level, itemStack);
+                org.bukkit.event.player.PlayerShearEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerShearEntityEvent(player, this, itemStack, hand, drops);
+                if (event.isCancelled()) return InteractionResult.PASS;
+                drops = org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getDrops());
+                this.shear(level, SoundSource.PLAYERS, itemStack, drops);
+                // Paper end - call PlayerShearEntityEvent
                 this.gameEvent(GameEvent.SHEAR, player);
                 itemStack.hurtAndBreak(1, player, hand.asEquipmentSlot());
             }
@@ -138,40 +_,33 @@
             }
 
             if (this.stewEffects != null) {
-                if (this.level() instanceof ServerLevel level) {
-                    level.sendParticles(
-                        ParticleTypes.SMOKE,
-                        this.getX(),
-                        this.getY(0.5),
-                        this.getZ(),
-                        2,
-                        0.5,
-                        0.0,
-                        0.5,
-                        0.0,
-                        0.2,
-                        0.0,
-                        ClientboundLevelParticlesPacket.RandomizationType.ALTERNATIVE_WITH_SPEED
-                    );
+                for (int i = 0; i < 2; i++) {
+                    this.level()
+                        .addParticle(
+                            ParticleTypes.SMOKE,
+                            this.getX() + this.random.nextDouble() / 2.0,
+                            this.getY(0.5),
+                            this.getZ() + this.random.nextDouble() / 2.0,
+                            0.0,
+                            this.random.nextDouble() / 5.0,
+                            0.0
+                        );
                 }
             } else {
                 itemStack.consume(1, player);
-                if (this.level() instanceof ServerLevel level) {
-                    SpellParticleOption particle = SpellParticleOption.create(ParticleTypes.EFFECT, -1, 1.0F);
-                    level.sendParticles(
-                        particle,
-                        this.getX(),
-                        this.getY(0.5),
-                        this.getZ(),
-                        4,
-                        0.5,
-                        0.0,
-                        0.5,
-                        0.0,
-                        0.2,
-                        0.0,
-                        ClientboundLevelParticlesPacket.RandomizationType.ALTERNATIVE_WITH_SPEED
-                    );
+                SpellParticleOption particle = SpellParticleOption.create(ParticleTypes.EFFECT, -1, 1.0F);
+
+                for (int i = 0; i < 4; i++) {
+                    this.level()
+                        .addParticle(
+                            particle,
+                            this.getX() + this.random.nextDouble() / 2.0,
+                            this.getY(0.5),
+                            this.getZ() + this.random.nextDouble() / 2.0,
+                            0.0,
+                            this.random.nextDouble() / 5.0,
+                            0.0
+                        );
                 }
 
                 this.stewEffects = effectsFromItemStack.get();
@@ -187,11 +_,26 @@
     @Override
     public void shear(ServerLevel level, SoundSource soundSource, ItemStack tool) {
         level.playSound(null, this, SoundEvents.MOOSHROOM_SHEAR, soundSource, 1.0F, 1.0F);
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityTypes.COW, (timer) -> {})) return;
         this.convertTo(EntityTypes.COW, ConversionParams.single(this, false, false), cow -> {
+            net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, cow);
             level.sendParticles(ParticleTypes.EXPLOSION, this.getX(), this.getY(0.5), this.getZ(), 1, 0.0, 0.0, 0.0, 0.0);
+            // Purpur start - Fix cow rotation when shearing mooshroom
+            cow.copyPosition(this);
+            cow.yBodyRot = this.yBodyRot;
+            cow.setYHeadRot(this.getYHeadRot());
+            cow.yRotO = this.yRotO;
+            cow.xRotO = this.xRotO;
+            // Purpur end - Fix cow rotation when shearing mooshroom
             this.dropFromShearingLootTable(level, BuiltInLootTables.SHEAR_MOOSHROOM, tool, (l, drop) -> {
                 for (int i = 0; i < drop.getCount(); i++) {
-                    l.addFreshEntity(new ItemEntity(this.level(), this.getX(), this.getY(1.0), this.getZ(), drop.copyWithCount(1)));
+                    // Neo: Change from addFreshEntity to spawnAtLocation to ensure captureDrops can capture this, we also need to unset the default pickup delay from the item
+                    // Vanilla uses this.getY(1.0) for the y-level, which is this.getY() + this.getBbHeight() * 1.0, so we pass the BB height as the Y-offset.
+                    ItemEntity item = spawnAtLocation(level, drop.copyWithCount(1), this.getBbHeight());
+                    if (item != null) {
+                        // addFreshEntity does not incur a pickup delay, while spawnAtLocation sets the default pickup delay.
+                        item.setNoPickUpDelay();
+                    }
                 }
             });
         });
