--- a/net/minecraft/server/commands/CommandSpawnpoint.java
+++ b/net/minecraft/server/commands/CommandSpawnpoint.java
@@ -42,7 +42,7 @@
         float f1 = vec2f.x;
 
         for (EntityPlayer entityplayer : collection) {
-            entityplayer.setRespawnPosition(new EntityPlayer.RespawnConfig(WorldData.a.of(resourcekey, blockposition, f, f1), true), false);
+            entityplayer.setRespawnPosition(new EntityPlayer.RespawnConfig(WorldData.a.of(resourcekey, blockposition, f, f1), true), false, org.bukkit.event.player.PlayerSpawnChangeEvent.Cause.COMMAND); // CraftBukkit
         }
 
         String s = resourcekey.location().toString();
