--- a/net/minecraft/server/commands/CommandGive.java
+++ b/net/minecraft/server/commands/CommandGive.java
@@ -53,7 +53,7 @@
                     boolean flag = entityplayer.getInventory().add(itemstack1);
 
                     if (flag && itemstack1.isEmpty()) {
-                        EntityItem entityitem = entityplayer.drop(itemstack, false);
+                        EntityItem entityitem = entityplayer.drop(itemstack, false, false, false); // CraftBukkit - SPIGOT-2942: Add boolean to call event
 
                         if (entityitem != null) {
                             entityitem.makeFakeItem();
