--- a/net/minecraft/server/commands/GameRuleCommand.java
+++ b/net/minecraft/server/commands/GameRuleCommand.java
@@ -42,7 +42,7 @@
         CommandSourceStack commandsourcestack = (CommandSourceStack) commandcontext.getSource();
         T t0 = (T) commandcontext.getArgument("value", gamerule.valueClass());
 
-        commandsourcestack.getLevel().getGameRules().set(gamerule, t0, ((CommandSourceStack) commandcontext.getSource()).getServer());
+        commandsourcestack.getLevel().getGameRules().set(gamerule, t0, ((CommandSourceStack) commandcontext.getSource()).getLevel()); // CraftBukkit
         commandsourcestack.sendSuccess(() -> {
             return Component.translatable("commands.gamerule.set", gamerule.id(), gamerule.serialize(t0));
         }, true);
