--- a/net/minecraft/server/commands/ScheduleCommand.java
+++ b/net/minecraft/server/commands/ScheduleCommand.java
@@ -58,7 +58,7 @@
         } else {
             long j = commandsourcestack.getLevel().getGameTime() + (long) i;
             Identifier identifier = (Identifier) pair.getFirst();
-            TimerQueue<MinecraftServer> timerqueue = commandsourcestack.getServer().getWorldData().overworldData().getScheduledEvents();
+            TimerQueue<MinecraftServer> timerqueue = commandsourcestack.getLevel().serverLevelData.overworldData().getScheduledEvents(); // CraftBukkit - SPIGOT-6667: Use world specific function timer
             Optional<CommandFunction<CommandSourceStack>> optional = ((Either) pair.getSecond()).left();
 
             if (optional.isPresent()) {
