--- a/net/minecraft/server/commands/WeatherCommand.java
+++ b/net/minecraft/server/commands/WeatherCommand.java
@@ -33,11 +33,11 @@
     }
 
     private static int getDuration(CommandSourceStack commandsourcestack, int i, IntProvider intprovider) {
-        return i == -1 ? intprovider.sample(commandsourcestack.getServer().overworld().getRandom()) : i;
+        return i == -1 ? intprovider.sample(commandsourcestack.getLevel().getRandom()) : i; // CraftBukkit - SPIGOT-7680: per-world
     }
 
     private static int setClear(CommandSourceStack commandsourcestack, int i) {
-        commandsourcestack.getServer().overworld().setWeatherParameters(getDuration(commandsourcestack, i, ServerLevel.RAIN_DELAY), 0, false, false);
+        commandsourcestack.getLevel().setWeatherParameters(getDuration(commandsourcestack, i, ServerLevel.RAIN_DELAY), 0, false, false); // CraftBukkit - SPIGOT-7680: per-world
         commandsourcestack.sendSuccess(() -> {
             return Component.translatable("commands.weather.set.clear");
         }, true);
@@ -45,7 +45,7 @@
     }
 
     private static int setRain(CommandSourceStack commandsourcestack, int i) {
-        commandsourcestack.getServer().overworld().setWeatherParameters(0, getDuration(commandsourcestack, i, ServerLevel.RAIN_DURATION), true, false);
+        commandsourcestack.getLevel().setWeatherParameters(0, getDuration(commandsourcestack, i, ServerLevel.RAIN_DURATION), true, false); // CraftBukkit - SPIGOT-7680: per-world
         commandsourcestack.sendSuccess(() -> {
             return Component.translatable("commands.weather.set.rain");
         }, true);
@@ -53,7 +53,7 @@
     }
 
     private static int setThunder(CommandSourceStack commandsourcestack, int i) {
-        commandsourcestack.getServer().overworld().setWeatherParameters(0, getDuration(commandsourcestack, i, ServerLevel.THUNDER_DURATION), true, true);
+        commandsourcestack.getLevel().setWeatherParameters(0, getDuration(commandsourcestack, i, ServerLevel.THUNDER_DURATION), true, true); // CraftBukkit - SPIGOT-7680: per-world
         commandsourcestack.sendSuccess(() -> {
             return Component.translatable("commands.weather.set.thunder");
         }, true);
