--- a/net/minecraft/world/item/BundleItem.java
+++ b/net/minecraft/world/item/BundleItem.java
@@ -9,7 +_,6 @@
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ARGB;
 import net.minecraft.util.Mth;
-import net.minecraft.util.Prediction;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.Entity;
@@ -59,7 +_,7 @@
     @Override
     public boolean overrideStackedOnOther(ItemStack self, Slot slot, ClickAction clickAction, Player player) {
         BundleContents initialContents = self.get(DataComponents.BUNDLE_CONTENTS);
-        if (initialContents == null) {
+        if (initialContents == null || self.getCount() != 1) {
             return false;
         }
 
@@ -96,6 +_,7 @@
 
     @Override
     public boolean overrideOtherStackedOnMe(ItemStack self, ItemStack other, Slot slot, ClickAction clickAction, Player player, SlotAccess carriedItem) {
+        if (self.getCount() != 1) return false;
         if (clickAction == ClickAction.PRIMARY && other.isEmpty()) {
             toggleSelectedItem(self, -1);
             return false;
@@ -193,7 +_,7 @@
         if (contents != null && !contents.isEmpty()) {
             Optional<ItemStack> itemStack = removeOneItemFromBundle(bundle, player, contents);
             if (itemStack.isPresent()) {
-                player.drop(itemStack.get(), true, Prediction.PREDICTED);
+                player.drop(itemStack.get(), true);
                 return true;
             } else {
                 return false;
