--- a/net/minecraft/world/item/CompassItem.java
+++ b/net/minecraft/world/item/CompassItem.java
@@ -8,7 +_,6 @@
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
-import net.minecraft.util.Prediction;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EquipmentSlot;
@@ -62,7 +_,7 @@
             itemStack.consume(1, player);
             lodestoneCompass.set(DataComponents.LODESTONE_TRACKER, target);
             if (!player.getInventory().add(lodestoneCompass)) {
-                player.drop(lodestoneCompass, false, Prediction.PREDICTED);
+                player.drop(lodestoneCompass, false);
             }
         }
 
