--- a/net/minecraft/world/item/CreativeModeTabs.java
+++ b/net/minecraft/world/item/CreativeModeTabs.java
@@ -502,8 +_,6 @@
                     coloredBlocks.accept(Items.TERRACOTTA);
                     registerColoredItems(coloredBlocks, gameplayColorOrder, Items.DYED_TERRACOTTA);
                     registerColoredItems(coloredBlocks, gameplayColorOrder, Items.CONCRETE);
-                    registerColoredItems(coloredBlocks, gameplayColorOrder, Items.CONCRETE_STAIRS);
-                    registerColoredItems(coloredBlocks, gameplayColorOrder, Items.CONCRETE_SLAB);
                     registerColoredItems(coloredBlocks, gameplayColorOrder, Items.CONCRETE_POWDER);
                     registerColoredItems(coloredBlocks, gameplayColorOrder, Items.GLAZED_TERRACOTTA);
                     coloredBlocks.accept(Items.GLASS);
@@ -915,7 +_,6 @@
                         registerColoredItems(functionalBlocks, gameplayColorOrder, Items.DYED_SHULKER_BOX);
                         functionalBlocks.accept(Items.RESPAWN_ANCHOR);
                         registerColoredItems(functionalBlocks, gameplayColorOrder, Items.BED);
-                        functionalBlocks.accept(Items.STRAW_BED);
                         functionalBlocks.accept(Items.CANDLE);
                         registerColoredItems(functionalBlocks, gameplayColorOrder, Items.DYED_CANDLE);
                         registerColoredItems(functionalBlocks, gameplayColorOrder, Items.BANNER);
@@ -931,6 +_,7 @@
                         functionalBlocks.accept(Items.END_PORTAL_FRAME);
                         functionalBlocks.accept(Items.VAULT);
                         functionalBlocks.accept(Items.ENDER_EYE);
+                        functionalBlocks.accept(Items.STRAW_BED);
                         Items.COPPER_GOLEM_STATUE.forEach(functionalBlocks::accept);
                         functionalBlocks.accept(Items.INFESTED_STONE);
                         functionalBlocks.accept(Items.INFESTED_COBBLESTONE);
