--- a/net/minecraft/world/item/CrossbowItem.java
+++ b/net/minecraft/world/item/CrossbowItem.java
@@ -45,7 +_,7 @@
     private static final float START_SOUND_PERCENT = 0.2F;
     private static final float MID_SOUND_PERCENT = 0.5F;
     private static final float ARROW_POWER = 3.15F;
-    private static final float FIREWORK_POWER = 1.6F;
+    public static final float FIREWORK_POWER = 1.6F;
     public static final float MOB_ARROW_POWER = 1.6F;
     private static final CrossbowItem.ChargingSounds DEFAULT_SOUNDS = new CrossbowItem.ChargingSounds(
         Optional.of(SoundEvents.CROSSBOW_LOADING_START), Optional.of(SoundEvents.CROSSBOW_LOADING_MIDDLE), Optional.of(SoundEvents.CROSSBOW_LOADING_END)
@@ -90,7 +_,14 @@
     public void releaseUsing(ItemStack p_40875_, Level p_40876_, LivingEntity p_40877_, int p_40878_) {
         int i = this.getUseDuration(p_40875_, p_40877_) - p_40878_;
         float f = getPowerForTime(i, p_40875_, p_40877_);
-        if (f >= 1.0F && !isCharged(p_40875_) && tryLoadProjectiles(p_40877_, p_40875_)) {
+        // Paper start - Add EntityLoadCrossbowEvent
+        if (f >= 1.0F && !isCharged(p_40875_)) {
+            final io.papermc.paper.event.entity.EntityLoadCrossbowEvent event = new io.papermc.paper.event.entity.EntityLoadCrossbowEvent(p_40877_.getBukkitLivingEntity(), p_40875_.asBukkitMirror(), org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(p_40877_.getUsedItemHand()));
+            if (!event.callEvent() || !tryLoadProjectiles(p_40877_, p_40875_, event.shouldConsumeItem()) || !event.shouldConsumeItem()) {
+                if (p_40877_ instanceof ServerPlayer player) player.containerMenu.sendAllDataToRemote();
+                return;
+            }
+            // Paper end - Add EntityLoadCrossbowEvent
             CrossbowItem.ChargingSounds crossbowitem$chargingsounds = this.getChargingSounds(p_40875_);
             crossbowitem$chargingsounds.end()
                 .ifPresent(
@@ -109,7 +_,11 @@
     }
 
     private static boolean tryLoadProjectiles(LivingEntity p_40860_, ItemStack p_40861_) {
-        List<ItemStack> list = draw(p_40861_, p_40860_.getProjectile(p_40861_), p_40860_);
+        // Paper start - Add EntityLoadCrossbowEvent
+        return CrossbowItem.tryLoadProjectiles(p_40860_, p_40861_, true);
+    }
+    private static boolean tryLoadProjectiles(LivingEntity p_40860_, ItemStack p_40861_, boolean consume) {
+        List<ItemStack> list = draw(p_40861_, p_40860_.getProjectile(p_40861_), p_40860_, consume);
         if (!list.isEmpty()) {
             p_40861_.set(DataComponents.CHARGED_PROJECTILES, ChargedProjectiles.of(list));
             return true;
@@ -161,7 +_,11 @@
     @Override
     protected Projectile createProjectile(Level p_331583_, LivingEntity p_40863_, ItemStack p_40864_, ItemStack p_40865_, boolean p_40866_) {
         if (p_40865_.is(Items.FIREWORK_ROCKET)) {
-            return new FireworkRocketEntity(p_331583_, p_40865_, p_40863_, p_40863_.getX(), p_40863_.getEyeY() - 0.15F, p_40863_.getZ(), true);
+            // Paper start
+            FireworkRocketEntity entity =  new FireworkRocketEntity(p_331583_, p_40865_, p_40863_, p_40863_.getX(), p_40863_.getEyeY() - 0.15F, p_40863_.getZ(), true);
+            entity.spawningEntity = p_40863_.getUUID(); // Paper
+            return entity;
+            // Paper end
         } else {
             Projectile projectile = super.createProjectile(p_331583_, p_40863_, p_40864_, p_40865_, p_40866_);
             if (projectile instanceof AbstractArrow abstractarrow) {
@@ -181,6 +_,7 @@
         Level p_40888_, LivingEntity p_40889_, InteractionHand p_40890_, ItemStack p_40891_, float p_40892_, float p_40893_, @Nullable LivingEntity p_331602_
     ) {
         if (p_40888_ instanceof ServerLevel serverlevel) {
+            if (p_40889_ instanceof Player player && net.neoforged.neoforge.event.EventHooks.onArrowLoose(p_40891_, p_40889_.level(), player, 1, true) < 0) return;
             ChargedProjectiles chargedprojectiles = p_40891_.set(DataComponents.CHARGED_PROJECTILES, ChargedProjectiles.EMPTY);
             if (chargedprojectiles != null && !chargedprojectiles.isEmpty()) {
                 this.shoot(serverlevel, p_40889_, p_40890_, p_40891_, chargedprojectiles.getItems(), p_40892_, p_40893_, p_40889_ instanceof Player, p_331602_);
