--- a/net/minecraft/world/item/CushionItem.java
+++ b/net/minecraft/world/item/CushionItem.java
@@ -42,7 +_,7 @@
         BlockPos blockPos = placeContext.getClickedPos();
         Vec3 entityPos = Vec3.atCenterOfWithY(blockPos, recalculatedContext.getClickLocation().y);
         AABB spawnAABB = EntityTypes.CUSHION.getSpawnAABB(entityPos);
-        if (!Cushion.canBePlacedAt(level, spawnAABB)) {
+        if (!Cushion.wouldSuriveAt(level, spawnAABB)) {
             return InteractionResult.FAIL;
         }
 
@@ -62,7 +_,7 @@
             serverLevel.addFreshEntity(cushion);
             cushion.destroyIfInFire(serverLevel);
             level.playSound(null, cushion.getX(), cushion.getY(), cushion.getZ(), SoundEvents.CUSHION_PLACE, SoundSource.BLOCKS, 0.75F, 0.8F);
-            cushion.gameEvent(GameEvent.ENTITY_PLACE, context.getPlayer());
+            cushion.gameEvent(GameEvent.ENTITY_PLACE);
             itemStack.consume(1, placeContext.getPlayer());
         }
 
