--- a/net/minecraft/world/item/DebugStickItem.java
+++ b/net/minecraft/world/item/DebugStickItem.java
@@ -48,8 +_,12 @@
         return InteractionResult.sidedSuccess(level.isClientSide);
     }
 
+    public boolean handleInteractionSpigot(Player p_150803_, BlockState p_150804_, LevelAccessor p_150805_, BlockPos p_150806_, boolean p_150807_, ItemStack p_150808_) {
+        return handleInteraction(p_150803_, p_150804_, p_150805_, p_150806_, p_150807_, p_150808_);
+    }
+
     private boolean handleInteraction(
-        Player p_150803_, BlockState p_150804_, LevelAccessor p_150805_, BlockPos p_150806_, boolean p_150807_, ItemStack p_150808_
+            Player p_150803_, BlockState p_150804_, LevelAccessor p_150805_, BlockPos p_150806_, boolean p_150807_, ItemStack p_150808_
     ) {
         if (!p_150803_.canUseGameMasterBlocks()) {
             return false;
