--- a/net/minecraft/world/item/DyeItem.java
+++ b/net/minecraft/world/item/DyeItem.java
@@ -10,7 +_,7 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.SignBlockEntity;
-import net.minecraft.world.level.block.entity.SignTextSlot;
+import org.bukkit.event.entity.SheepDyeWoolEvent;
 
 public class DyeItem extends Item implements SignApplicator {
     public DyeItem(Item.Properties properties) {
@@ -24,7 +_,17 @@
             if (dyeColor != null && sheep.getColor() != dyeColor) {
                 sheep.level().playSound(player, sheep, SoundEvents.DYE_USE, SoundSource.PLAYERS, 1.0F, 1.0F);
                 if (!player.level().isClientSide()) {
-                    sheep.setColor(dyeColor);
+                    // CraftBukkit start
+                    byte bColor = (byte) dyeColor.getId();
+                    org.bukkit.event.entity.SheepDyeWoolEvent event = new org.bukkit.event.entity.SheepDyeWoolEvent((org.bukkit.entity.Sheep) sheep.getBukkitEntity(), org.bukkit.DyeColor.getByWoolData(bColor), (org.bukkit.entity.Player) player.getBukkitEntity());
+                    sheep.level().getCraftServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        return InteractionResult.PASS;
+                    }
+
+                    sheep.setColor(DyeColor.byId((byte) event.getColor().getWoolData()));
+                    // CraftBukkit end
                     itemStack.shrink(1);
                 }
 
@@ -36,9 +_,9 @@
     }
 
     @Override
-    public boolean tryApplyToSign(Level level, SignBlockEntity sign, SignTextSlot slot, ItemStack item, Player player) {
+    public boolean tryApplyToSign(Level level, SignBlockEntity sign, boolean isFrontText, ItemStack item, Player player) {
         DyeColor dye = item.get(DataComponents.DYE);
-        if (dye != null && sign.updateText(text -> text.withColor(dye), slot)) {
+        if (dye != null && sign.updateText(text -> text.setColor(dye), isFrontText)) {
             level.playSound(null, sign.getBlockPos(), SoundEvents.DYE_USE, SoundSource.BLOCKS, 1.0F, 1.0F);
             return true;
         } else {
