--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -86,6 +_,11 @@
                     eyeofender.signalTo(blockpos);
                     p_41184_.gameEvent(GameEvent.PROJECTILE_SHOOT, eyeofender.position(), GameEvent.Context.of(p_41185_));
                     p_41184_.addFreshEntity(eyeofender);
+                    // CraftBukkit start
+                    if (!p_41184_.canAddFreshEntity()) {
+                        return InteractionResultHolder.fail(itemstack);
+                    }
+                    // CraftBukkit end
                     if (p_41185_ instanceof ServerPlayer serverplayer) {
                         CriteriaTriggers.USED_ENDER_EYE.trigger(serverplayer, blockpos);
                     }
