--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -90,7 +_,11 @@
                 eyeOfEnder.setItem(itemStack);
                 eyeOfEnder.signalTo(Vec3.atLowerCornerOf(nearestMapFeature));
                 level.gameEvent(GameEvent.PROJECTILE_SHOOT, eyeOfEnder.position(), GameEvent.Context.of(player));
-                level.addFreshEntity(eyeOfEnder);
+                // CraftBukkit start
+                if (!level.addFreshEntity(eyeOfEnder)) {
+                    return InteractionResult.FAIL;
+                }
+                // CraftBukkit end
                 if (player instanceof ServerPlayer serverPlayer) {
                     CriteriaTriggers.USED_ENDER_EYE.trigger(serverPlayer, nearestMapFeature);
                 }
