--- a/net/minecraft/world/item/FireChargeItem.java
+++ b/net/minecraft/world/item/FireChargeItem.java
@@ -3,7 +_,6 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Position;
-import net.minecraft.core.dispenser.BlockSource;
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.util.RandomSource;
@@ -13,14 +_,11 @@
 import net.minecraft.world.item.context.UseOnContext;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BaseFireBlock;
-import net.minecraft.world.level.block.CampfireBlock;
-import net.minecraft.world.level.block.CandleBlock;
-import net.minecraft.world.level.block.CandleCakeBlock;
 import net.minecraft.world.level.block.DispenserBlock;
 import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class FireChargeItem extends Item implements ProjectileItem {
     public FireChargeItem(Item.Properties p_41202_) {
@@ -33,17 +_,34 @@
         BlockPos blockpos = p_41204_.getClickedPos();
         BlockState blockstate = level.getBlockState(blockpos);
         boolean flag = false;
-        if (!CampfireBlock.canLight(blockstate) && !CandleBlock.canLight(blockstate) && !CandleCakeBlock.canLight(blockstate)) {
+        BlockState blockstate2 = blockstate.getToolModifiedState(p_41204_, net.neoforged.neoforge.common.ItemAbilities.FIRESTARTER_LIGHT, false);
+        if (blockstate2 == null) {
             blockpos = blockpos.relative(p_41204_.getClickedFace());
             if (BaseFireBlock.canBePlacedAt(level, blockpos, p_41204_.getHorizontalDirection())) {
+                // CraftBukkit start - fire BlockIgniteEvent // Youer add null check
+                if (p_41204_.getPlayer() != null && CraftEventFactory.callBlockIgniteEvent(level, blockpos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FIREBALL, p_41204_.getPlayer()).isCancelled()) {
+                    if (!p_41204_.getPlayer().getAbilities().instabuild) {
+                        p_41204_.getItemInHand().shrink(1);
+                    }
+                    return InteractionResult.PASS;
+                }
+                // CraftBukkit end
                 this.playSound(level, blockpos);
                 level.setBlockAndUpdate(blockpos, BaseFireBlock.getState(level, blockpos));
                 level.gameEvent(p_41204_.getPlayer(), GameEvent.BLOCK_PLACE, blockpos);
                 flag = true;
             }
         } else {
+            // CraftBukkit start - fire BlockIgniteEvent // Youer add null check
+            if (p_41204_.getPlayer() != null &&  CraftEventFactory.callBlockIgniteEvent(level, blockpos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FIREBALL, p_41204_.getPlayer()).isCancelled()) {
+                if (!p_41204_.getPlayer().getAbilities().instabuild) {
+                    p_41204_.getItemInHand().shrink(1);
+                }
+                return InteractionResult.PASS;
+            }
+            // CraftBukkit end
             this.playSound(level, blockpos);
-            level.setBlockAndUpdate(blockpos, blockstate.setValue(BlockStateProperties.LIT, Boolean.valueOf(true)));
+            level.setBlockAndUpdate(blockpos, blockstate2);
             level.gameEvent(p_41204_.getPlayer(), GameEvent.BLOCK_CHANGE, blockpos);
             flag = true;
         }
@@ -87,5 +_,10 @@
             .power(1.0F)
             .overrideDispenseEvent(1018)
             .build();
+    }
+
+    @Override
+    public boolean canPerformAction(ItemStack stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_FIRECHARGE_ACTIONS.contains(itemAbility);
     }
 }
