--- a/net/minecraft/world/item/FireworkRocketItem.java
+++ b/net/minecraft/world/item/FireworkRocketItem.java
@@ -15,6 +_,7 @@
 import net.minecraft.world.item.context.UseOnContext;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityUnleashEvent;
 
 public class FireworkRocketItem extends Item implements ProjectileItem {
     public static final byte[] CRAFTABLE_DURATIONS = new byte[]{1, 2, 3};
@@ -59,7 +_,7 @@
         if (player.isFallFlying()) {
             ItemStack itemStack = player.getItemInHand(hand);
             if (level instanceof ServerLevel serverLevel) {
-                if (player.dropAllLeashConnections(null)) {
+                if (player.dropAllLeashConnections(null, EntityUnleashEvent.UnleashReason.FIREWORK)) { // CraftBukkit
                     level.playSound(null, player, SoundEvents.LEAD_BREAK, SoundSource.NEUTRAL, 1.0F, 1.0F);
                 }
 
