--- a/net/minecraft/world/item/FireworkRocketItem.java
+++ b/net/minecraft/world/item/FireworkRocketItem.java
@@ -42,8 +_,12 @@
                 vec3.z + (double)direction.getStepZ() * 0.15,
                 itemstack
             );
-            level.addFreshEntity(fireworkrocketentity);
-            itemstack.shrink(1);
+            // Paper start - PlayerLaunchProjectileEvent
+            io.papermc.paper.event.player.PlayerLaunchProjectileEvent event = new io.papermc.paper.event.player.PlayerLaunchProjectileEvent((org.bukkit.entity.Player) p_41216_.getPlayer().getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), (org.bukkit.entity.Firework) fireworkrocketentity.getBukkitEntity());
+            if (!event.callEvent() || !level.addFreshEntity(fireworkrocketentity)) return InteractionResult.PASS;
+            if (event.shouldConsume() && !p_41216_.getPlayer().getAbilities().instabuild) itemstack.shrink(1);
+            else if (p_41216_.getPlayer() instanceof net.minecraft.server.level.ServerPlayer) ((net.minecraft.server.level.ServerPlayer) p_41216_.getPlayer()).getBukkitEntity().updateInventory();
+            // Paper end - PlayerLaunchProjectileEvent
         }
 
         return InteractionResult.sidedSuccess(level.isClientSide);
