--- a/net/minecraft/world/item/HoneycombItem.java
+++ b/net/minecraft/world/item/HoneycombItem.java
@@ -15,8 +_,6 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.data.recipes.RecipeCategory;
 import net.minecraft.server.level.ServerPlayer;
-import net.minecraft.sounds.SoundEvents;
-import net.minecraft.sounds.SoundSource;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.context.UseOnContext;
@@ -32,6 +_,10 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 
 public class HoneycombItem extends Item implements SignApplicator {
+    /**
+     * @deprecated Neo: Use the {@link net.neoforged.neoforge.registries.datamaps.builtin.NeoForgeDataMaps#WAXABLES data map}, this field will be ignored in a future version
+     */
+    @Deprecated
     public static final Supplier<BiMap<Block, Block>> WAXABLES = Suppliers.memoize(
         () -> {
             Builder<Block, Block> builderx = ImmutableBiMap.builder();
@@ -56,6 +_,10 @@
             return builderx.build();
         }
     );
+    /**
+     * @deprecated Use the {@link net.neoforged.neoforge.common.DataMapHooks#INVERSE_WAXABLES_DATAMAP inverse map} generated from the data map, this field will be ignored in a future version
+     */
+    @Deprecated
     public static final Supplier<BiMap<Block, Block>> WAX_OFF_BY_BLOCK = Suppliers.memoize(() -> WAXABLES.get().inverse());
     public static final ImmutableMap<Block, Pair<RecipeCategory, String>> WAXED_RECIPES;
 
@@ -71,6 +_,14 @@
         return getWaxed(oldState).map(waxedState -> {
             Player player = context.getPlayer();
             ItemStack itemInHand = context.getItemInHand();
+            // Paper start - EntityChangeBlockEvent
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(player, pos, waxedState)) {
+                if (!player.isCreative()) {
+                    player.containerMenu.forceHeldSlot(context.getHand());
+                }
+                return InteractionResult.PASS;
+            }
+            // Paper end
             if (player instanceof ServerPlayer serverPlayer) {
                 CriteriaTriggers.ITEM_USED_ON_BLOCK.trigger(serverPlayer, pos, itemInHand);
             }
@@ -79,12 +_,10 @@
             level.setBlock(pos, waxedState, 11);
             level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(player, waxedState));
             level.levelEvent(player, 3003, pos, 0);
-            level.playSound(player, pos, SoundEvents.HONEYCOMB_WAX_ON, SoundSource.BLOCKS, 1.0F, 1.0F);
             if (oldState.getBlock() instanceof ChestBlock && oldState.getValue(ChestBlock.TYPE) != ChestType.SINGLE) {
                 BlockPos neighborPos = ChestBlock.getConnectedBlockPos(pos, oldState);
                 level.gameEvent(GameEvent.BLOCK_CHANGE, neighborPos, GameEvent.Context.of(player, level.getBlockState(neighborPos)));
                 level.levelEvent(player, 3003, neighborPos, 0);
-                level.playSound(player, pos, SoundEvents.HONEYCOMB_WAX_ON, SoundSource.BLOCKS, 1.0F, 1.0F);
             }
 
             return (InteractionResult)InteractionResult.SUCCESS;
@@ -92,15 +_,13 @@
     }
 
     public static Optional<BlockState> getWaxed(BlockState oldState) {
-        return Optional.ofNullable(WAXABLES.get().get(oldState.getBlock())).map(b -> b.withPropertiesOf(oldState));
+        return Optional.ofNullable(net.neoforged.neoforge.common.DataMapHooks.getBlockWaxed(oldState.getBlock())).map(b -> b.withPropertiesOf(oldState));
     }
 
     @Override
     public boolean tryApplyToSign(Level level, SignBlockEntity sign, boolean isFrontText, ItemStack item, Player player) {
         if (sign.setWaxed(true)) {
-            BlockPos blockPos = sign.getBlockPos();
-            level.levelEvent(null, 3003, blockPos, 0);
-            level.playSound(player, blockPos, SoundEvents.HONEYCOMB_WAX_ON, SoundSource.BLOCKS, 1.0F, 1.0F);
+            level.levelEvent(null, 3003, sign.getBlockPos(), 0);
             return true;
         } else {
             return false;
