--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.item;
 
 import com.google.common.collect.Lists;
+import com.mohistmc.youer.feature.ban.bans.BanItem;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.DataResult;
@@ -44,8 +_,6 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ExtraCodecs;
-import net.minecraft.util.Mth;
-import net.minecraft.util.NullOps;
 import net.minecraft.util.StringUtil;
 import net.minecraft.util.Unit;
 import net.minecraft.world.InteractionHand;
@@ -91,12 +_,13 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.Spawner;
 import net.minecraft.world.level.block.entity.BlockEntityType;
-import net.minecraft.world.level.block.entity.SignText;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.pattern.BlockInWorld;
 import net.minecraft.world.level.gameevent.GameEvent;
 import org.apache.commons.lang3.function.TriConsumer;
 import org.apache.commons.lang3.mutable.MutableBoolean;
+import org.bukkit.event.player.PlayerItemDamageEvent;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -189,9 +_,17 @@
                 if (itemStack.isEmpty()) {
                     output.writeVarInt(0);
                 } else {
-                    output.writeVarInt(itemStack.getCount());
+                    output.writeVarInt(io.papermc.paper.util.sanitizer.ItemComponentSanitizer.sanitizeCount(io.papermc.paper.util.sanitizer.ItemObfuscationSession.currentSession(), itemStack, itemStack.getCount())); // Paper - potentially sanitize count
                     Item.STREAM_CODEC.encode(output, itemStack.typeHolder());
-                    patchCodec.encode(output, itemStack.components.asPatch());
+                    // Paper start - adventure; conditionally render translatable components
+                    boolean prev = net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
+                    try (final io.papermc.paper.util.SafeAutoClosable ignored = io.papermc.paper.util.sanitizer.ItemObfuscationSession.withContext(c -> c.itemStack(itemStack))) { // pass the itemstack as context to the obfuscation session
+                        net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(true);
+                        patchCodec.encode(output, itemStack.components.asPatch());
+                    } finally {
+                        net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(prev);
+                    }
+                    // Paper end - adventure; conditionally render translatable components
                 }
             }
         };
@@ -202,7 +_,7 @@
             public ItemStack decode(RegistryFriendlyByteBuf input) {
                 ItemStack itemStack = codec.decode(input);
                 if (!itemStack.isEmpty()) {
-                    RegistryOps<Unit> ops = input.registryAccess().createSerializationContext(NullOps.INSTANCE);
+                    RegistryOps<io.papermc.paper.util.CountingOps.Value> ops = input.registryAccess().createSerializationContext(io.papermc.paper.util.CountingOps.INSTANCE); // Paper - Track codec depth
                     ItemStack.CODEC.encodeStart(ops, itemStack).getOrThrow(DecoderException::new);
                 }
 
@@ -240,6 +_,10 @@
         return !this.isEmpty() && this.components.hasNonDefault(type);
     }
 
+    public boolean isComponentsPatchEmpty() {
+        return this.isEmpty() || this.components.isPatchEmpty();
+    }
+
     public ItemStack(ItemLike item, int count) {
         this(item.asItem().builtInRegistryHolder(), count);
     }
@@ -359,14 +_,38 @@
     }
 
     public InteractionResult useOn(UseOnContext context) {
+        if (BanItem.checkMoShou(context.getPlayer(), context.getItemInHand())) {
+            context.getPlayer().setItemInHand(context.getHand(), ItemStack.EMPTY);
+            return InteractionResult.FAIL;
+        }
+        if (BanItem.check(context.getPlayer(), context.getItemInHand())) return InteractionResult.FAIL;
+        var e = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent(context, net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent.UsePhase.ITEM_AFTER_BLOCK));
+        if (e.isCanceled()) return e.getCancellationResult();
+        if (!context.getLevel().isClientSide()) return net.neoforged.neoforge.common.CommonHooks.onPlaceItemIntoWorld(context);
+        return onItemUse(context, (c) -> getItem().useOn(context));
+    }
+
+    @Override
+    public InteractionResult onItemUseFirst(UseOnContext context) {
+        var e = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent(context, net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent.UsePhase.ITEM_BEFORE_BLOCK));
+        if (e.isCanceled()) return e.getCancellationResult();
+        return onItemUse(context, (c) -> getItem().onItemUseFirst(this, context));
+    }
+
+    private InteractionResult onItemUse(UseOnContext context, java.util.function.Function<UseOnContext, InteractionResult> callback) {
         Player player = context.getPlayer();
+        if (BanItem.checkMoShou(context.getPlayer(), context.getItemInHand())) {
+            context.getPlayer().setItemInHand(context.getHand(), ItemStack.EMPTY);
+            return InteractionResult.FAIL;
+        }
+        if (BanItem.check(player, context.getItemInHand())) return InteractionResult.FAIL;
         BlockPos pos = context.getClickedPos();
         if (player != null && !player.getAbilities().mayBuild && !this.canPlaceOnBlockInAdventureMode(new BlockInWorld(context.getLevel(), pos, false))) {
             return InteractionResult.PASS;
         }
 
         Item usedItem = this.getItem();
-        InteractionResult result = usedItem.useOn(context);
+        InteractionResult result = callback.apply(context);
         if (player != null && result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
             player.awardStat(Stats.ITEM_USED.get(usedItem));
         }
@@ -413,6 +_,12 @@
         return result;
     }
 
+    // Neo: Delegate max stack size logic to the item to allow it to be stack-sensitive
+    @Override
+    public int getMaxStackSize() {
+        return this.getItem().getMaxStackSize(this);
+    }
+
     public boolean isStackable() {
         return this.getMaxStackSize() > 1 && (!this.isDamageableItem() || !this.isDamaged());
     }
@@ -422,19 +_,39 @@
     }
 
     public boolean isDamaged() {
-        return this.isDamageableItem() && this.getDamageValue() > 0;
-    }
+        return this.isDamageableItem() && getItem().isDamaged(this);
+    }
+
+    // Purpur start - Add option to mend the most damaged equipment first
+    public float getDamagePercent() {
+        if (this.has(DataComponents.UNBREAKABLE)) {
+            return 0.0F;
+        }
+
+        final int maxDamage = this.getOrDefault(DataComponents.MAX_DAMAGE, 0);
+        if (maxDamage == 0) {
+            return 0.0F;
+        }
+
+        final int damage = this.getOrDefault(DataComponents.DAMAGE, 0);
+        if (damage == 0) {
+            return 0.0F;
+        }
+
+        return (float) damage / maxDamage;
+    }
+    // Purpur end - Add option to mend the most damaged equipment first
 
     public int getDamageValue() {
-        return Mth.clamp(this.getOrDefault(DataComponents.DAMAGE, 0), 0, this.getMaxDamage());
+        return this.getItem().getDamage(this);
     }
 
     public void setDamageValue(int value) {
-        this.set(DataComponents.DAMAGE, Mth.clamp(value, 0, this.getMaxDamage()));
+        this.getItem().setDamage(this, value);
     }
 
     public int getMaxDamage() {
-        return this.getOrDefault(DataComponents.MAX_DAMAGE, 0);
+        return this.getItem().getMaxDamage(this);
     }
 
     public boolean isBroken() {
@@ -445,17 +_,48 @@
         return this.isDamageableItem() && this.getDamageValue() >= this.getMaxDamage() - 1;
     }
 
-    public void hurtAndBreak(int amount, ServerLevel level, @Nullable ServerPlayer player, Consumer<Item> onBreak) {
+    public void hurtAndBreak(int amount, ServerLevel level, @Nullable LivingEntity player, Consumer<Item> onBreak) {
+        // Paper start - add force boolean overload
+        this.hurtAndBreak(amount, level, player, onBreak, false);
+    }
+
+    public void hurtAndBreak(int amount, ServerLevel level, @Nullable LivingEntity player, Consumer<Item> onBreak, final boolean force) { // Paper - Add EntityDamageItemEvent
+        // Paper end
+        amount = getItem().damageItem(this, amount, player, onBreak);
         int newAmount = this.processDurabilityChange(amount, level, player);
-        if (newAmount != 0) {
+        // CraftBukkit start
+        if (player instanceof ServerPlayer serverplayer) {
+            PlayerItemDamageEvent event = new PlayerItemDamageEvent(serverplayer.getBukkitEntity(), CraftItemStack.asCraftMirror(this), newAmount);
+            event.getPlayer().getServer().getPluginManager().callEvent(event);
+
+            if (newAmount != event.getDamage() || event.isCancelled()) {
+                event.getPlayer().updateInventory();
+            }
+            if (event.isCancelled()) {
+                return;
+            }
+
+            newAmount = event.getDamage();
+        }
+        // CraftBukkit end
+        if (newAmount != 0) { // Paper - Add EntityDamageItemEvent - diff on change for above event ifs.
             this.applyDamage(this.getDamageValue() + newAmount, player, onBreak);
         }
     }
 
     private int processDurabilityChange(int amount, ServerLevel level, @Nullable ServerPlayer player) {
+        return processDurabilityChange(amount, level, (LivingEntity) player);
+    }
+
+    private int processDurabilityChange(int amount, ServerLevel level, @Nullable LivingEntity player) {
+        // Paper start - itemstack damage api
+        return this.processDurabilityChange(amount, level, player, false);
+    }
+    private int processDurabilityChange(final int amount, final ServerLevel level, final @Nullable LivingEntity player, final boolean force) {
+        // Paper end - itemstack damage api
         if (!this.isDamageableItem()) {
             return 0;
-        } else if (player != null && player.hasInfiniteMaterials()) {
+        } else if (player instanceof ServerPlayer && player.hasInfiniteMaterials() && !force) { // Paper - Add EntityDamageItemEvent
             return 0;
         } else {
             return amount > 0 ? EnchantmentHelper.processDurabilityChange(level, this, amount) : amount;
@@ -463,13 +_,28 @@
     }
 
     private void applyDamage(int newDamage, @Nullable ServerPlayer player, Consumer<Item> onBreak) {
-        if (player != null) {
-            CriteriaTriggers.ITEM_DURABILITY_CHANGED.trigger(player, this, newDamage);
+        applyDamage(newDamage, (LivingEntity) player, onBreak);
+    }
+
+    private void applyDamage(int newDamage, @Nullable LivingEntity player, Consumer<Item> onBreak) {
+        if (player instanceof ServerPlayer serverPlayer) {
+            CriteriaTriggers.ITEM_DURABILITY_CHANGED.trigger(serverPlayer, this, newDamage); // Paper - Add EntityDamageItemEvent
         }
 
         this.setDamageValue(newDamage);
         if (this.isBroken()) {
             Item item = this.getItem();
+            // CraftBukkit start - Check for item breaking
+            if (this.count == 1 && player instanceof ServerPlayer serverplayer) {
+                org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerItemBreakEvent(serverplayer, this);
+            }
+            // CraftBukkit end
+            // Purpur start - Implement elytra settings
+            if (this.has(DataComponents.GLIDER)) {
+                setDamageValue(this.getMaxDamage() - 1);
+                return;
+            }
+            // Purpur end - Implement elytra settings
             this.shrink(1);
             onBreak.accept(item);
         }
@@ -482,7 +_,25 @@
                 return;
             }
 
-            int newDamage = Math.min(this.getDamageValue() + newAmount, this.getMaxDamage() - 1);
+            int newDamage = Math.min(this.getDamageValue() + newAmount, this.getMaxDamage() - 1); // Paper - Expand PlayerItemDamageEvent - diff on change as min computation is copied post event.
+
+            // Paper start - Expand PlayerItemDamageEvent
+            if (newDamage - this.getDamageValue() > 0) {
+                org.bukkit.event.player.PlayerItemDamageEvent event = new org.bukkit.event.player.PlayerItemDamageEvent(
+                        serverPlayer.getBukkitEntity(),
+                        org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(this),
+                        newDamage - this.getDamageValue(),
+                        amount
+                );
+                if (!event.callEvent() || event.getDamage() == 0) {
+                    return;
+                }
+
+                // Prevent breaking the item in this code path as callers may expect the item to survive
+                // (given the method name)
+                newDamage = Math.min(this.getDamageValue() + event.getDamage(), this.getMaxDamage() - 1);
+            }
+            // Paper end - Expand PlayerItemDamageEvent
             this.applyDamage(newDamage, serverPlayer, i -> {});
         }
     }
@@ -492,9 +_,14 @@
     }
 
     public void hurtAndBreak(int amount, LivingEntity owner, EquipmentSlot slot) {
+        // Paper start - add param to skip infinite mats check
+        this.hurtAndBreak(amount, owner, slot, false);
+    }
+    public void hurtAndBreak(final int amount, final LivingEntity owner, final EquipmentSlot slot, final boolean force) {
+        // Paper end - add param to skip infinite mats check
         if (owner.level() instanceof ServerLevel serverLevel) {
             this.hurtAndBreak(
-                amount, serverLevel, owner instanceof ServerPlayer player ? player : null, brokenItem -> owner.onEquippedItemBroken(brokenItem, slot)
+                    amount, serverLevel, owner, brokenItem -> {if (slot != null) owner.onEquippedItemBroken(brokenItem, slot); }, force // Paper - Add EntityDamageItemEvent & itemstack damage API - do not process entity related callbacks when damaging from API
             );
         }
     }
@@ -579,11 +_,17 @@
     }
 
     public ItemStack copy() {
-        if (this.isEmpty()) {
+        // Paper start - Perf: Optimize Hoppers
+        return this.copy(false);
+    }
+
+    public ItemStack copy(final boolean originalItem) {
+        if (!originalItem && this.isEmpty()) {
+            // Paper end - Perf: Optimize Hoppers
             return EMPTY;
         }
 
-        ItemStack copy = new ItemStack(this.typeHolder(), this.count, this.components.copy());
+        ItemStack copy = new ItemStack(originalItem ? this.item : this.typeHolder(), this.count, this.components.copy()); // Paper - Perf: Optimize Hoppers
         copy.setPopTime(this.getPopTime());
         return copy;
     }
@@ -618,6 +_,16 @@
         }
     }
 
+    /**
+     * Compares an itemstack with an {@link ItemStackTemplate} as per {@link #matches(ItemStack, ItemStack)}.
+     */
+    public static boolean matches(ItemStack a, @Nullable ItemStackTemplate b) {
+        if (b == null) {
+            return a.isEmpty();
+        }
+        return a.getCount() == b.count() && isSameItemSameComponents(a, b);
+    }
+
     @Deprecated
     public static boolean listMatches(List<ItemStack> left, List<ItemStack> right) {
         if (left.size() != right.size()) {
@@ -645,6 +_,27 @@
         }
     }
 
+
+    /**
+     * {@return true if a and b refer to the same item, or if a is empty and b is null}
+     */
+    public static boolean isSameItem(ItemStack a, @Nullable ItemStackTemplate b) {
+        return b == null ? a.isEmpty() : a.is(b.item());
+    }
+
+    /**
+     * Compares the item and components of this stack against an {@link ItemStackTemplate}.
+     *
+     * @return True if either this stack is empty and the template is null, or they reference the same item and have equivalent component patches.
+     */
+    public static boolean isSameItemSameComponents(ItemStack a, @Nullable ItemStackTemplate b) {
+        if (a.isEmpty() || b == null) {
+            return a.isEmpty() == (b == null);
+        } else {
+            return a.is(b.item()) && a.components.patchEquals(b.components());
+        }
+    }
+
     public static boolean matchesIgnoringComponents(ItemStack a, ItemStack b, Predicate<DataComponentType<?>> ignoredPredicate) {
         if (a == b) {
             return true;
@@ -759,6 +_,12 @@
         return this.getItem().useOnRelease(this);
     }
 
+    // CraftBukkit start
+    public void restorePatch(DataComponentPatch datacomponentpatch) {
+        this.components.restorePatch(datacomponentpatch);
+    }
+    // CraftBukkit end
+
     public <T> @Nullable T set(DataComponentType<T> type, @Nullable T value) {
         return this.components.set(type, value);
     }
@@ -802,6 +_,25 @@
         this.components.setAll(components);
     }
 
+    // Paper start - (this is just a good no conflict location)
+    public org.bukkit.inventory.ItemStack asBukkitMirror() {
+        return CraftItemStack.asCraftMirror(this);
+    }
+    public org.bukkit.inventory.ItemStack asBukkitCopy() {
+        return CraftItemStack.asCraftMirror(this.copy());
+    }
+    public static ItemStack fromBukkitCopy(org.bukkit.inventory.ItemStack itemstack) {
+        return CraftItemStack.asNMSCopy(itemstack);
+    }
+    private CraftItemStack bukkitStack;
+    public org.bukkit.inventory.ItemStack getBukkitStack() {
+        if (bukkitStack == null || bukkitStack.handle != this) {
+            bukkitStack = CraftItemStack.asCraftMirror(this);
+        }
+        return bukkitStack;
+    }
+    // Paper end
+
     public Component getHoverName() {
         Component customName = this.getCustomName();
         return customName != null ? customName : this.getItemName();
@@ -829,7 +_,7 @@
     }
 
     public Component getStyledHoverName() {
-        MutableComponent hoverName = Component.empty().append(this.getHoverName()).withStyle(this.getRarity().color());
+        MutableComponent hoverName = Component.empty().append(this.getHoverName()).withStyle(this.getRarity().getStyleModifier());
         if (this.has(DataComponents.CUSTOM_NAME)) {
             hoverName.withStyle(ChatFormatting.ITALIC);
         }
@@ -837,24 +_,14 @@
         return hoverName;
     }
 
-    public <T> void addToTooltip(
-        DataComponentType<T> type,
-        TooltipProvider.Getter<T> tooltipGetter,
-        Item.TooltipContext context,
-        TooltipDisplay display,
-        Consumer<Component> consumer,
-        TooltipFlag flag
-    ) {
-        T component = this.get(type);
-        if (component != null && display.shows(type)) {
-            tooltipGetter.get(component).addToTooltip(context, consumer, flag, this.components);
-        }
-    }
-
+    @Override // Neo: Add override annotation to ensure our extension method signature always matches vanilla
     public <T extends TooltipProvider> void addToTooltip(
         DataComponentType<T> type, Item.TooltipContext context, TooltipDisplay display, Consumer<Component> consumer, TooltipFlag flag
     ) {
-        this.addToTooltip(type, c -> c, context, display, consumer, flag);
+        T component = (T)this.get(type);
+        if (component != null && display.shows(type)) {
+            component.addToTooltip(context, consumer, flag, this.components);
+        }
     }
 
     public List<Component> getTooltipLines(Item.TooltipContext context, @Nullable Player player, TooltipFlag tooltipFlag) {
@@ -866,6 +_,7 @@
             List<Component> lines = Lists.newArrayList();
             lines.add(this.getStyledHoverName());
             this.addDetailsToTooltip(context, display, player, tooltipFlag, lines::add);
+            net.neoforged.neoforge.event.EventHooks.onItemTooltip(this, player, lines, tooltipFlag, context);
             return lines;
         }
     }
@@ -895,11 +_,10 @@
         this.addToTooltip(DataComponents.PROFILE, context, display, builder, tooltipFlag);
         this.addToTooltip(DataComponents.LORE, context, display, builder, tooltipFlag);
         this.addToTooltip(DataComponents.SULFUR_CUBE_CONTENT, context, display, builder, tooltipFlag);
-        this.addAttributeTooltips(builder, display, player);
+        // Neo: Replace attribute tooltips with custom handling
+        net.neoforged.neoforge.common.util.AttributeUtil.addAttributeTooltips(this, builder, display, net.neoforged.neoforge.common.util.AttributeTooltipContext.of(player, context, display, tooltipFlag));
         this.addUnitComponentToTooltip(DataComponents.INTANGIBLE_PROJECTILE, INTANGIBLE_TOOLTIP, display, builder);
         this.addUnitComponentToTooltip(DataComponents.UNBREAKABLE, UNBREAKABLE_TOOLTIP, display, builder);
-        this.addToTooltip(DataComponents.SIGN_TEXT_FRONT, SignText.FRONT_TEXT, context, display, builder, tooltipFlag);
-        this.addToTooltip(DataComponents.SIGN_TEXT_BACK, SignText.BACK_TEXT, context, display, builder, tooltipFlag);
         this.addToTooltip(DataComponents.OMINOUS_BOTTLE_AMPLIFIER, context, display, builder, tooltipFlag);
         this.addToTooltip(DataComponents.SUSPICIOUS_STEW_EFFECTS, context, display, builder, tooltipFlag);
         this.addToTooltip(DataComponents.BLOCK_STATE, context, display, builder, tooltipFlag);
@@ -950,7 +_,10 @@
             builder.accept(component);
         }
     }
-
+    /**
+     * @deprecated Neo: Use {@link net.neoforged.neoforge.common.util.AttributeUtil#addAttributeTooltips}
+     */
+    @Deprecated
     private void addAttributeTooltips(Consumer<Component> consumer, TooltipDisplay display, @Nullable Player player) {
         if (display.shows(DataComponents.ATTRIBUTE_MODIFIERS)) {
             for (EquipmentSlotGroup slot : EquipmentSlotGroup.values()) {
@@ -1005,22 +_,65 @@
         return !this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY).isEmpty();
     }
 
+    // Purpur start - Config to allow unsafe enchants
+    public boolean hasEnchantment(Holder<Enchantment> enchantment) {
+        return this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY).getLevel(enchantment) > 0;
+    }
+    // Purpur end - Config to allow unsafe enchants
+
+    /**
+     * Gets all enchantments from NBT. Use {@link ItemStack#getAllEnchantments} for gameplay logic.
+     */
+    @Override
+    public ItemEnchantments getTagEnchantments() {
+        return getEnchantments();
+    }
+
+    /**
+     * @deprecated Neo: Use {@link #getTagEnchantments()} for NBT enchantments, or {@link #getAllEnchantments} for gameplay.
+     */
+    @Deprecated
     public ItemEnchantments getEnchantments() {
         return this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
     }
 
     public void forEachModifier(EquipmentSlotGroup slot, TriConsumer<Holder<Attribute>, AttributeModifier, ItemAttributeModifiers.Display> consumer) {
+        // Neo: Reflect real attribute modifiers when doing iteration
+        this.getAttributeModifiers().forEach(slot, consumer);
+        if (false) {
+        // Start disabled vanilla code
         ItemAttributeModifiers modifiers = this.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
         modifiers.forEach(slot, consumer);
+        // end disabled vanilla code
+        }
         EnchantmentHelper.forEachModifier(this, slot, (a, b) -> consumer.accept(a, b, ItemAttributeModifiers.Display.attributeModifiers()));
     }
 
     public void forEachModifier(EquipmentSlot slot, BiConsumer<Holder<Attribute>, AttributeModifier> consumer) {
+        // Neo: Reflect real attribute modifiers when doing iteration
+        this.getAttributeModifiers().forEach(slot, consumer);
+        if (false) {
+        // Start disabled vanilla code
         ItemAttributeModifiers modifiers = this.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
         modifiers.forEach(slot, consumer);
+        // end disabled vanilla code
+        }
         EnchantmentHelper.forEachModifier(this, slot, consumer);
     }
 
+    // CraftBukkit start
+    @Deprecated
+    public void setItem(Item item) {
+        this.bukkitStack = null;
+        this.item = item.builtInRegistryHolder();
+        // Paper start - change base component prototype
+        final DataComponentPatch patch = this.getComponentsPatch();
+        this.components = new PatchedDataComponentMap(this.item.components());
+        this.applyComponents(patch);
+        // Paper end - change base component prototype
+    }
+    // CraftBukkit end
+
     public Component getDisplayName() {
         MutableComponent hoverName = Component.empty().append(this.getHoverName());
         if (this.has(DataComponents.CUSTOM_NAME)) {
@@ -1029,7 +_,7 @@
 
         MutableComponent result = ComponentUtils.wrapInSquareBrackets(hoverName);
         if (!this.isEmpty()) {
-            result.withStyle(this.getRarity().color()).withStyle(s -> s.withHoverEvent(new HoverEvent.ShowItem(ItemStackTemplate.fromNonEmptyStack(this))));
+            result.withStyle(this.getRarity().getStyleModifier()).withStyle(s -> s.withHoverEvent(new HoverEvent.ShowItem(ItemStackTemplate.fromNonEmptyStack(this))));
         }
 
         return result;
@@ -1085,7 +_,7 @@
     }
 
     public void consume(int amount, @Nullable LivingEntity owner) {
-        if (owner == null || !owner.hasInfiniteMaterials()) {
+        if ((owner == null || !owner.hasInfiniteMaterials())) {
             this.shrink(amount);
         }
     }
@@ -1110,11 +_,14 @@
         }
     }
 
+    /** @deprecated Forge: Use {@linkplain net.neoforged.neoforge.common.extensions.IItemStackExtension#onDestroyed(ItemEntity, net.minecraft.world.damagesource.DamageSource) damage source sensitive version} */
+    @Deprecated
     public void onDestroyed(ItemEntity itemEntity) {
         this.getItem().onDestroyed(itemEntity);
     }
 
     public boolean canBeHurtBy(DamageSource source) {
+        if (!getItem().canBeHurtBy(this, source)) return false;
         DamageResistant damageResistant = this.get(DataComponents.DAMAGE_RESISTANT);
         return damageResistant == null || !damageResistant.isResistantTo(source);
     }
