--- a/net/minecraft/world/item/ItemUtils.java
+++ b/net/minecraft/world/item/ItemUtils.java
@@ -1,7 +_,6 @@
 package net.minecraft.world.item;
 
 import java.util.stream.Stream;
-import net.minecraft.util.Prediction;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.item.ItemEntity;
@@ -29,7 +_,7 @@
             }
 
             if (!player.getInventory().add(newItemStack)) {
-                player.drop(newItemStack, false, Prediction.PREDICTED);
+                player.drop(newItemStack, false);
             }
 
             return itemStack;
@@ -43,7 +_,15 @@
     public static void onContainerDestroyed(ItemEntity container, Stream<ItemStack> contents) {
         Level level = container.level();
         if (!level.isClientSide()) {
-            contents.forEach(stack -> level.addFreshEntity(new ItemEntity(level, container.getX(), container.getY(), container.getZ(), stack)));
+            // Paper start - call EntityDropItemEvent
+            contents.forEach(stack -> {
+                ItemEntity droppedItem = new ItemEntity(level, container.getX(), container.getY(), container.getZ(), stack);
+                org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(container.getBukkitEntity(), (org.bukkit.entity.Item) droppedItem.getBukkitEntity());
+                if (event.callEvent()) {
+                    level.addFreshEntity(droppedItem);
+                }
+            });
+            // Paper end - call EntityDropItemEvent
         }
     }
 }
