--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -16,6 +_,7 @@
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.function.Predicate;
+import java.util.function.Supplier;
 import java.util.function.UnaryOperator;
 import net.minecraft.ChatFormatting;
 import net.minecraft.advancements.triggers.CriteriaTriggers;
@@ -96,6 +_,32 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import org.apache.commons.lang3.function.TriConsumer;
 import org.apache.commons.lang3.mutable.MutableBoolean;
+// CraftBukkit start
+import java.util.Map;
+import net.minecraft.core.Direction;
+import net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket;
+import net.minecraft.sounds.SoundSource;
+import net.minecraft.world.level.block.BedBlock;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.SaplingBlock;
+import net.minecraft.world.level.block.SignBlock;
+import net.minecraft.world.level.block.BaseEntityBlock;
+import net.minecraft.world.level.block.WitherSkullBlock;
+import net.minecraft.world.level.block.SoundType;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.SignBlockEntity;
+import net.minecraft.world.level.block.entity.SkullBlockEntity;
+import org.bukkit.Location;
+import org.bukkit.TreeType;
+import org.bukkit.craftbukkit.block.CapturedBlockState;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.block.CraftBlockState;
+import org.bukkit.craftbukkit.util.CraftLocation;
+import org.bukkit.event.block.BlockFertilizeEvent;
+import org.bukkit.event.player.PlayerItemDamageEvent;
+import org.bukkit.event.world.StructureGrowEvent;
+// CraftBukkit end
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -177,18 +_,28 @@
                 int count = input.readVarInt();
                 if (count <= 0) {
                     return ItemStack.EMPTY;
+                } else {
+                    Holder<Item> item = Item.STREAM_CODEC.decode(input);
+                    DataComponentPatch patch = patchCodec.decode(input);
+                    // CraftBukkit start
+                    ItemStack itemstack = new ItemStack(item, count, patch);
+                    if (!patch.isEmpty()) {
+                        CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack));
+                    }
+                    return itemstack;
+                    // CraftBukkit end
                 }
-
-                Holder<Item> item = Item.STREAM_CODEC.decode(input);
-                DataComponentPatch patch = patchCodec.decode(input);
-                return new ItemStack(item, count, patch);
             }
 
             public void encode(RegistryFriendlyByteBuf output, ItemStack itemStack) {
-                if (itemStack.isEmpty()) {
+                if (itemStack.isEmpty() || itemStack.getItem() == null) { // CraftBukkit - NPE fix itemstack.getItem()
                     output.writeVarInt(0);
                 } else {
                     output.writeVarInt(itemStack.getCount());
+                    // Spigot start - filter
+                    itemStack = itemStack.copy();
+                    CraftItemStack.setItemMeta(itemStack, CraftItemStack.getItemMeta(itemStack));
+                    // Spigot end
                     Item.STREAM_CODEC.encode(output, itemStack.typeHolder());
                     patchCodec.encode(output, itemStack.components.asPatch());
                 }
@@ -239,6 +_,10 @@
         return !this.isEmpty() && this.components.hasNonDefault(type);
     }
 
+    public boolean isComponentsPatchEmpty() {
+        return this.isEmpty() || this.components.isPatchEmpty();
+    }
+
     public ItemStack(ItemLike item, int count) {
         this(item.asItem().builtInRegistryHolder(), count);
     }
@@ -358,6 +_,20 @@
     }
 
     public InteractionResult useOn(UseOnContext context) {
+        var e = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent(context, net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent.UsePhase.ITEM_AFTER_BLOCK));
+        if (e.isCanceled()) return e.getCancellationResult();
+        if (!context.getLevel().isClientSide()) return net.neoforged.neoforge.common.CommonHooks.onPlaceItemIntoWorld(context);
+        return onItemUse(context, (c) -> getItem().useOn(context));
+    }
+
+    @Override
+    public InteractionResult onItemUseFirst(UseOnContext context) {
+        var e = net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent(context, net.neoforged.neoforge.event.entity.player.UseItemOnBlockEvent.UsePhase.ITEM_BEFORE_BLOCK));
+        if (e.isCanceled()) return e.getCancellationResult();
+        return onItemUse(context, (c) -> getItem().onItemUseFirst(this, context));
+    }
+
+    private InteractionResult onItemUse(UseOnContext context, java.util.function.Function<UseOnContext, InteractionResult> callback) {
         Player player = context.getPlayer();
         BlockPos pos = context.getClickedPos();
         if (player != null && !player.getAbilities().mayBuild && !this.canPlaceOnBlockInAdventureMode(new BlockInWorld(context.getLevel(), pos, false))) {
@@ -365,7 +_,7 @@
         }
 
         Item usedItem = this.getItem();
-        InteractionResult result = usedItem.useOn(context);
+        InteractionResult result = callback.apply(context);
         if (player != null && result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
             player.awardStat(Stats.ITEM_USED.get(usedItem));
         }
@@ -412,6 +_,12 @@
         return result;
     }
 
+    // Neo: Delegate max stack size logic to the item to allow it to be stack-sensitive
+    @Override
+    public int getMaxStackSize() {
+        return this.getItem().getMaxStackSize(this);
+    }
+
     public boolean isStackable() {
         return this.getMaxStackSize() > 1 && (!this.isDamageableItem() || !this.isDamaged());
     }
@@ -421,19 +_,19 @@
     }
 
     public boolean isDamaged() {
-        return this.isDamageableItem() && this.getDamageValue() > 0;
+        return this.isDamageableItem() && getItem().isDamaged(this);
     }
 
     public int getDamageValue() {
-        return Mth.clamp(this.getOrDefault(DataComponents.DAMAGE, 0), 0, this.getMaxDamage());
+        return this.getItem().getDamage(this);
     }
 
     public void setDamageValue(int value) {
-        this.set(DataComponents.DAMAGE, Mth.clamp(value, 0, this.getMaxDamage()));
+        this.getItem().setDamage(this, value);
     }
 
     public int getMaxDamage() {
-        return this.getOrDefault(DataComponents.MAX_DAMAGE, 0);
+        return this.getItem().getMaxDamage(this);
     }
 
     public boolean isBroken() {
@@ -444,14 +_,40 @@
         return this.isDamageableItem() && this.getDamageValue() >= this.getMaxDamage() - 1;
     }
 
-    public void hurtAndBreak(int amount, ServerLevel level, @Nullable ServerPlayer player, Consumer<Item> onBreak) {
+    public void hurtAndBreak(int amount, ServerLevel level, @Nullable LivingEntity player, Consumer<Item> onBreak) {
+        // Paper start - add force boolean overload
+        this.hurtAndBreak(amount, level, player, onBreak, false);
+    }
+
+    public void hurtAndBreak(int amount, ServerLevel level, @Nullable LivingEntity player, Consumer<Item> onBreak, final boolean force) { // Paper - Add EntityDamageItemEvent
+        // Paper end
+        amount = getItem().damageItem(this, amount, player, onBreak);
         int newAmount = this.processDurabilityChange(amount, level, player);
+        // CraftBukkit start
+        if (player instanceof ServerPlayer serverplayer) {
+            PlayerItemDamageEvent event = new PlayerItemDamageEvent(serverplayer.getBukkitEntity(), CraftItemStack.asCraftMirror(this), newAmount);
+            event.getPlayer().getServer().getPluginManager().callEvent(event);
+
+            if (newAmount != event.getDamage() || event.isCancelled()) {
+                event.getPlayer().updateInventory();
+            }
+            if (event.isCancelled()) {
+                return;
+            }
+
+            newAmount = event.getDamage();
+        }
+        // CraftBukkit end
         if (newAmount != 0) {
             this.applyDamage(this.getDamageValue() + newAmount, player, onBreak);
         }
     }
 
     private int processDurabilityChange(int amount, ServerLevel level, @Nullable ServerPlayer player) {
+        return processDurabilityChange(amount, level, (LivingEntity) player);
+    }
+
+    private int processDurabilityChange(int amount, ServerLevel level, @Nullable LivingEntity player) {
         if (!this.isDamageableItem()) {
             return 0;
         } else if (player != null && player.hasInfiniteMaterials()) {
@@ -462,13 +_,22 @@
     }
 
     private void applyDamage(int newDamage, @Nullable ServerPlayer player, Consumer<Item> onBreak) {
-        if (player != null) {
-            CriteriaTriggers.ITEM_DURABILITY_CHANGED.trigger(player, this, newDamage);
+        applyDamage(newDamage, (LivingEntity) player, onBreak);
+    }
+
+    private void applyDamage(int newDamage, @Nullable LivingEntity player, Consumer<Item> onBreak) {
+        if (player instanceof ServerPlayer serverPlayer) {
+            CriteriaTriggers.ITEM_DURABILITY_CHANGED.trigger(serverPlayer, this, newDamage);
         }
 
         this.setDamageValue(newDamage);
         if (this.isBroken()) {
             Item item = this.getItem();
+            // CraftBukkit start - Check for item breaking
+            if (this.count == 1 && player instanceof ServerPlayer serverplayer) {
+                org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerItemBreakEvent(serverplayer, this);
+            }
+            // CraftBukkit end
             this.shrink(1);
             onBreak.accept(item);
         }
@@ -491,9 +_,14 @@
     }
 
     public void hurtAndBreak(int amount, LivingEntity owner, EquipmentSlot slot) {
+        // Paper start - add param to skip infinite mats check
+        this.hurtAndBreak(amount, owner, slot, false);
+    }
+    public void hurtAndBreak(final int amount, final LivingEntity owner, final EquipmentSlot slot, final boolean force) {
+        // Paper end - add param to skip infinite mats check
         if (owner.level() instanceof ServerLevel serverLevel) {
             this.hurtAndBreak(
-                amount, serverLevel, owner instanceof ServerPlayer player ? player : null, brokenItem -> owner.onEquippedItemBroken(brokenItem, slot)
+                    amount, serverLevel, owner, brokenItem -> {if (slot != null) owner.onEquippedItemBroken(brokenItem, slot); }, force // Paper - Add EntityDamageItemEvent & itemstack damage API - do not process entity related callbacks when damaging from API
             );
         }
     }
@@ -578,11 +_,17 @@
     }
 
     public ItemStack copy() {
-        if (this.isEmpty()) {
+        // Paper start - Perf: Optimize Hoppers
+        return this.copy(false);
+    }
+
+    public ItemStack copy(final boolean originalItem) {
+        if (!originalItem && this.isEmpty()) {
+            // Paper end - Perf: Optimize Hoppers
             return EMPTY;
         }
 
-        ItemStack copy = new ItemStack(this.typeHolder(), this.count, this.components.copy());
+        ItemStack copy = new ItemStack(originalItem ? this.item : this.typeHolder(), this.count, this.components.copy()); // Paper - Perf: Optimize Hoppers
         copy.setPopTime(this.getPopTime());
         return copy;
     }
@@ -617,6 +_,16 @@
         }
     }
 
+    /**
+     * Compares an itemstack with an {@link ItemStackTemplate} as per {@link #matches(ItemStack, ItemStack)}.
+     */
+    public static boolean matches(ItemStack a, @Nullable ItemStackTemplate b) {
+        if (b == null) {
+            return a.isEmpty();
+        }
+        return a.getCount() == b.count() && isSameItemSameComponents(a, b);
+    }
+
     @Deprecated
     public static boolean listMatches(List<ItemStack> left, List<ItemStack> right) {
         if (left.size() != right.size()) {
@@ -644,6 +_,27 @@
         }
     }
 
+
+    /**
+     * {@return true if a and b refer to the same item, or if a is empty and b is null}
+     */
+    public static boolean isSameItem(ItemStack a, @Nullable ItemStackTemplate b) {
+        return b == null ? a.isEmpty() : a.is(b.item());
+    }
+
+    /**
+     * Compares the item and components of this stack against an {@link ItemStackTemplate}.
+     *
+     * @return True if either this stack is empty and the template is null, or they reference the same item and have equivalent component patches.
+     */
+    public static boolean isSameItemSameComponents(ItemStack a, @Nullable ItemStackTemplate b) {
+        if (a.isEmpty() || b == null) {
+            return a.isEmpty() == (b == null);
+        } else {
+            return a.is(b.item()) && a.components.patchEquals(b.components());
+        }
+    }
+
     public static boolean matchesIgnoringComponents(ItemStack a, ItemStack b, Predicate<DataComponentType<?>> ignoredPredicate) {
         if (a == b) {
             return true;
@@ -758,6 +_,12 @@
         return this.getItem().useOnRelease(this);
     }
 
+    // CraftBukkit start
+    public void restorePatch(DataComponentPatch datacomponentpatch) {
+        this.components.restorePatch(datacomponentpatch);
+    }
+    // CraftBukkit end
+
     public <T> @Nullable T set(DataComponentType<T> type, @Nullable T value) {
         return this.components.set(type, value);
     }
@@ -792,6 +_,25 @@
             this.components.restorePatch(oldPatch);
         }
     }
+    // Paper start - (this is just a good no conflict location)
+    public org.bukkit.inventory.ItemStack asBukkitMirror() {
+        return CraftItemStack.asCraftMirror(this);
+    }
+    public org.bukkit.inventory.ItemStack asBukkitCopy() {
+        return CraftItemStack.asCraftMirror(this.copy());
+    }
+    public static ItemStack fromBukkitCopy(org.bukkit.inventory.ItemStack itemstack) {
+        return CraftItemStack.asNMSCopy(itemstack);
+    }
+    private CraftItemStack bukkitStack;
+    public org.bukkit.inventory.ItemStack getBukkitStack() {
+        if (bukkitStack == null || bukkitStack.handle != this) {
+            bukkitStack = CraftItemStack.asCraftMirror(this);
+        }
+        return bukkitStack;
+    }
+    // Paper end
+
 
     public void applyComponents(DataComponentPatch patch) {
         this.components.applyPatch(patch);
@@ -828,7 +_,7 @@
     }
 
     public Component getStyledHoverName() {
-        MutableComponent hoverName = Component.empty().append(this.getHoverName()).withStyle(this.getRarity().color());
+        MutableComponent hoverName = Component.empty().append(this.getHoverName()).withStyle(this.getRarity().getStyleModifier());
         if (this.has(DataComponents.CUSTOM_NAME)) {
             hoverName.withStyle(ChatFormatting.ITALIC);
         }
@@ -836,6 +_,7 @@
         return hoverName;
     }
 
+    @Override // Neo: Add override annotation to ensure our extension method signature always matches vanilla
     public <T extends TooltipProvider> void addToTooltip(
         DataComponentType<T> type, Item.TooltipContext context, TooltipDisplay display, Consumer<Component> consumer, TooltipFlag flag
     ) {
@@ -854,6 +_,7 @@
             List<Component> lines = Lists.newArrayList();
             lines.add(this.getStyledHoverName());
             this.addDetailsToTooltip(context, display, player, tooltipFlag, lines::add);
+            net.neoforged.neoforge.event.EventHooks.onItemTooltip(this, player, lines, tooltipFlag, context);
             return lines;
         }
     }
@@ -883,7 +_,8 @@
         this.addToTooltip(DataComponents.PROFILE, context, display, builder, tooltipFlag);
         this.addToTooltip(DataComponents.LORE, context, display, builder, tooltipFlag);
         this.addToTooltip(DataComponents.SULFUR_CUBE_CONTENT, context, display, builder, tooltipFlag);
-        this.addAttributeTooltips(builder, display, player);
+        // Neo: Replace attribute tooltips with custom handling
+        net.neoforged.neoforge.common.util.AttributeUtil.addAttributeTooltips(this, builder, display, net.neoforged.neoforge.common.util.AttributeTooltipContext.of(player, context, display, tooltipFlag));
         this.addUnitComponentToTooltip(DataComponents.INTANGIBLE_PROJECTILE, INTANGIBLE_TOOLTIP, display, builder);
         this.addUnitComponentToTooltip(DataComponents.UNBREAKABLE, UNBREAKABLE_TOOLTIP, display, builder);
         this.addToTooltip(DataComponents.OMINOUS_BOTTLE_AMPLIFIER, context, display, builder, tooltipFlag);
@@ -936,7 +_,10 @@
             builder.accept(component);
         }
     }
-
+    /**
+     * @deprecated Neo: Use {@link net.neoforged.neoforge.common.util.AttributeUtil#addAttributeTooltips}
+     */
+    @Deprecated
     private void addAttributeTooltips(Consumer<Component> consumer, TooltipDisplay display, @Nullable Player player) {
         if (display.shows(DataComponents.ATTRIBUTE_MODIFIERS)) {
             for (EquipmentSlotGroup slot : EquipmentSlotGroup.values()) {
@@ -991,22 +_,59 @@
         return !this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY).isEmpty();
     }
 
+    /**
+     * Gets all enchantments from NBT. Use {@link ItemStack#getAllEnchantments} for gameplay logic.
+     */
+    @Override
+    public ItemEnchantments getTagEnchantments() {
+        return getEnchantments();
+    }
+
+    /**
+     * @deprecated Neo: Use {@link #getTagEnchantments()} for NBT enchantments, or {@link #getAllEnchantments} for gameplay.
+     */
+    @Deprecated
     public ItemEnchantments getEnchantments() {
         return this.getOrDefault(DataComponents.ENCHANTMENTS, ItemEnchantments.EMPTY);
     }
 
     public void forEachModifier(EquipmentSlotGroup slot, TriConsumer<Holder<Attribute>, AttributeModifier, ItemAttributeModifiers.Display> consumer) {
+        // Neo: Reflect real attribute modifiers when doing iteration
+        this.getAttributeModifiers().forEach(slot, consumer);
+        if (false) {
+        // Start disabled vanilla code
         ItemAttributeModifiers modifiers = this.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
         modifiers.forEach(slot, consumer);
+        // end disabled vanilla code
+        }
         EnchantmentHelper.forEachModifier(this, slot, (a, b) -> consumer.accept(a, b, ItemAttributeModifiers.Display.attributeModifiers()));
     }
 
     public void forEachModifier(EquipmentSlot slot, BiConsumer<Holder<Attribute>, AttributeModifier> consumer) {
+        // Neo: Reflect real attribute modifiers when doing iteration
+        this.getAttributeModifiers().forEach(slot, consumer);
+        if (false) {
+        // Start disabled vanilla code
         ItemAttributeModifiers modifiers = this.getOrDefault(DataComponents.ATTRIBUTE_MODIFIERS, ItemAttributeModifiers.EMPTY);
         modifiers.forEach(slot, consumer);
+        // end disabled vanilla code
+        }
         EnchantmentHelper.forEachModifier(this, slot, consumer);
     }
 
+    // CraftBukkit start
+    @Deprecated
+    public void setItem(Item item) {
+        this.bukkitStack = null;
+        this.item = item.builtInRegistryHolder();
+        // Paper start - change base component prototype
+        final DataComponentPatch patch = this.getComponentsPatch();
+        this.components = new PatchedDataComponentMap(this.item.components());
+        this.applyComponents(patch);
+        // Paper end - change base component prototype
+    }
+    // CraftBukkit end
+
     public Component getDisplayName() {
         MutableComponent hoverName = Component.empty().append(this.getHoverName());
         if (this.has(DataComponents.CUSTOM_NAME)) {
@@ -1015,7 +_,7 @@
 
         MutableComponent result = ComponentUtils.wrapInSquareBrackets(hoverName);
         if (!this.isEmpty()) {
-            result.withStyle(this.getRarity().color()).withStyle(s -> s.withHoverEvent(new HoverEvent.ShowItem(ItemStackTemplate.fromNonEmptyStack(this))));
+            result.withStyle(this.getRarity().getStyleModifier()).withStyle(s -> s.withHoverEvent(new HoverEvent.ShowItem(ItemStackTemplate.fromNonEmptyStack(this))));
         }
 
         return result;
@@ -1071,7 +_,7 @@
     }
 
     public void consume(int amount, @Nullable LivingEntity owner) {
-        if (owner == null || !owner.hasInfiniteMaterials()) {
+        if ((owner == null || !owner.hasInfiniteMaterials()) && this != ItemStack.EMPTY) { // CraftBukkit
             this.shrink(amount);
         }
     }
@@ -1096,11 +_,14 @@
         }
     }
 
+    /** @deprecated Forge: Use {@linkplain net.neoforged.neoforge.common.extensions.IItemStackExtension#onDestroyed(ItemEntity, net.minecraft.world.damagesource.DamageSource) damage source sensitive version} */
+    @Deprecated
     public void onDestroyed(ItemEntity itemEntity) {
         this.getItem().onDestroyed(itemEntity);
     }
 
     public boolean canBeHurtBy(DamageSource source) {
+        if (!getItem().canBeHurtBy(this, source)) return false;
         DamageResistant damageResistant = this.get(DataComponents.DAMAGE_RESISTANT);
         return damageResistant == null || !damageResistant.isResistantTo(source);
     }
