--- a/net/minecraft/world/item/JukeboxSongPlayer.java
+++ b/net/minecraft/world/item/JukeboxSongPlayer.java
@@ -13,9 +_,9 @@
 
 public class JukeboxSongPlayer {
     public static final int PLAY_EVENT_INTERVAL_TICKS = 20;
-    public long ticksSinceSongStarted;
+    private long ticksSinceSongStarted;
     @Nullable
-    public Holder<JukeboxSong> song;
+    private Holder<JukeboxSong> song;
     private final BlockPos blockPos;
     private final JukeboxSongPlayer.OnSongChanged onSongChanged;
 
