--- a/net/minecraft/world/item/JukeboxSong.java
+++ b/net/minecraft/world/item/JukeboxSong.java
@@ -6,12 +_,12 @@
 import net.minecraft.core.Holder;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentSerialization;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.Mth;
@@ -37,7 +_,7 @@
         JukeboxSong::comparatorOutput,
         JukeboxSong::new
     );
-    public static final Codec<Holder<JukeboxSong>> CODEC = RegistryCodecs.holder(Registries.JUKEBOX_SONG);
+    public static final Codec<Holder<JukeboxSong>> CODEC = RegistryFixedCodec.create(Registries.JUKEBOX_SONG);
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<JukeboxSong>> STREAM_CODEC = ByteBufCodecs.holder(
         Registries.JUKEBOX_SONG, DIRECT_STREAM_CODEC
     );
