--- a/net/minecraft/world/item/LingeringPotionItem.java
+++ b/net/minecraft/world/item/LingeringPotionItem.java
@@ -24,16 +_,23 @@
 
     @Override
     public InteractionResultHolder<ItemStack> use(Level p_42843_, Player p_42844_, InteractionHand p_42845_) {
-        p_42843_.playSound(
-            null,
-            p_42844_.getX(),
-            p_42844_.getY(),
-            p_42844_.getZ(),
-            SoundEvents.LINGERING_POTION_THROW,
-            SoundSource.NEUTRAL,
-            0.5F,
-            0.4F / (p_42843_.getRandom().nextFloat() * 0.4F + 0.8F)
-        );
-        return super.use(p_42843_, p_42844_, p_42845_);
+        // Paper start - PlayerLaunchProjectileEvent
+        InteractionResultHolder<ItemStack> wrapper = super.use(p_42843_, p_42844_, p_42845_);
+        if (wrapper.getResult() != net.minecraft.world.InteractionResult.FAIL) {
+            // Paper end - PlayerLaunchProjectileEvent
+            p_42843_.playSound(
+                    null,
+                    p_42844_.getX(),
+                    p_42844_.getY(),
+                    p_42844_.getZ(),
+                    SoundEvents.LINGERING_POTION_THROW,
+                    SoundSource.NEUTRAL,
+                    0.5F,
+                    0.4F / (p_42843_.getRandom().nextFloat() * 0.4F + 0.8F)
+            );
+            // Paper start - PlayerLaunchProjectileEvent
+        }
+        return wrapper;
+        // Paper end - PlayerLaunchProjectileEvent
     }
 }
