--- a/net/minecraft/world/item/MaceItem.java
+++ b/net/minecraft/world/item/MaceItem.java
@@ -22,6 +_,7 @@
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.util.CraftVector;
 import org.jspecify.annotations.Nullable;
 
 public class MaceItem extends Item {
@@ -123,7 +_,14 @@
             double knockbackPower = getKnockbackPower(attacker, nearby, direction);
             Vec3 knockbackVector = direction.normalize().scale(knockbackPower);
             if (knockbackPower > 0.0) {
-                nearby.push(knockbackVector.x, 0.7F, knockbackVector.z);
+                // CraftBukkit start - EntityKnockbackEvent
+                Vec3 vec3Push = new Vec3(knockbackVector.x, 0.7F, knockbackVector.z);
+                Vec3 result = attacker.getDeltaMovement().add(vec3Push);
+                org.bukkit.event.entity.EntityKnockbackEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityKnockbackEvent((org.bukkit.craftbukkit.entity.CraftLivingEntity) nearby.getBukkitEntity(), attacker, org.bukkit.event.entity.EntityKnockbackEvent.KnockbackCause.ENTITY_ATTACK, knockbackPower, result, vec3Push.x, vec3Push.y, vec3Push.z);
+                if (!event.isCancelled()) {
+                    nearby.push(CraftVector.toNMS(event.getFinalKnockback()));
+                }
+                // CraftBukkit end
                 if (nearby instanceof ServerPlayer otherPlayer) {
                     otherPlayer.connection.send(new ClientboundSetEntityMotionPacket(otherPlayer));
                 }
