--- a/net/minecraft/world/item/MaceItem.java
+++ b/net/minecraft/world/item/MaceItem.java
@@ -22,6 +_,7 @@
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.util.CraftVector;
 import org.jspecify.annotations.Nullable;
 
 public class MaceItem extends Item {
@@ -123,7 +_,7 @@
             double knockbackPower = getKnockbackPower(attacker, nearby, direction);
             Vec3 knockbackVector = direction.normalize().scale(knockbackPower);
             if (knockbackPower > 0.0) {
-                nearby.push(knockbackVector.x, 0.7F, knockbackVector.z);
+                nearby.push(knockbackVector.x, 0.7F, knockbackVector.z, attacker); // Paper - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
                 if (nearby instanceof ServerPlayer otherPlayer) {
                     otherPlayer.connection.send(new ClientboundSetEntityMotionPacket(otherPlayer));
                 }
