--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -52,11 +_,21 @@
     }
 
     public static @Nullable MapItemSavedData getSavedData(ItemStack itemStack, Level level) {
+        // Neo: Add instance method so that mods can override
+        Item map = itemStack.getItem();
+        if(map instanceof MapItem) {
+            return ((MapItem)map).getCustomMapData(itemStack, level);
+        }
+        return null;
+    }
+
+    @Nullable
+    protected MapItemSavedData getCustomMapData(ItemStack itemStack, Level level) {
         MapId id = itemStack.get(DataComponents.MAP_ID);
         return getSavedData(id, level);
     }
 
-    public static MapId createNewSavedData(
+    private static MapId createNewSavedData(
         ServerLevel level, int xSpawn, int zSpawn, int scale, boolean trackingPosition, boolean unlimitedTracking, ResourceKey<Level> dimension
     ) {
         MapItemSavedData newData = MapItemSavedData.createFresh(xSpawn, zSpawn, (byte)scale, trackingPosition, unlimitedTracking, dimension);
