--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -57,11 +_,21 @@
 
     @Nullable
     public static MapItemSavedData getSavedData(ItemStack p_42854_, Level p_42855_) {
+        // Neo: Add instance method so that mods can override
+        Item map = p_42854_.getItem();
+        if(map instanceof MapItem) {
+            return ((MapItem)map).getCustomMapData(p_42854_, p_42855_);
+        }
+        return null;
+    }
+
+    @Nullable
+    protected MapItemSavedData getCustomMapData(ItemStack p_42854_, Level p_42855_) {
         MapId mapid = p_42854_.get(DataComponents.MAP_ID);
         return getSavedData(mapid, p_42855_);
     }
 
-    public static MapId createNewSavedData(
+    private static MapId createNewSavedData(
         Level p_151121_, int p_151122_, int p_151123_, int p_151124_, boolean p_151125_, boolean p_151126_, ResourceKey<Level> p_151127_
     ) {
         MapItemSavedData mapitemsaveddata = MapItemSavedData.createFresh((double)p_151122_, (double)p_151123_, (byte)p_151124_, p_151125_, p_151126_, p_151127_);
