--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -58,12 +_,22 @@
     }
 
     public static @Nullable MapItemSavedData getSavedData(ItemStack itemStack, Level level) {
+        // Neo: Add instance method so that mods can override
+        Item map = itemStack.getItem();
+        if(map instanceof MapItem) {
+            return ((MapItem)map).getCustomMapData(itemStack, level);
+        }
+        return null;
+    }
+
+    @Nullable
+    protected MapItemSavedData getCustomMapData(ItemStack itemStack, Level level) {
         MapId id = itemStack.get(DataComponents.MAP_ID);
         return getSavedData(id, level);
     }
 
     public static MapId createNewSavedData(
-        ServerLevel level, int xSpawn, int zSpawn, int scale, boolean trackingPosition, boolean unlimitedTracking, ResourceKey<Level> dimension
+            ServerLevel level, int xSpawn, int zSpawn, int scale, boolean trackingPosition, boolean unlimitedTracking, ResourceKey<Level> dimension
     ) {
         MapItemSavedData newData = MapItemSavedData.createFresh(xSpawn, zSpawn, (byte)scale, trackingPosition, unlimitedTracking, dimension);
         MapId id = level.getFreeMapId();
@@ -101,10 +_,10 @@
                             int averagingAreaMinX = (centerX / scale + imgX - 64) * scale;
                             int averagingAreaMinZ = (centerZ / scale + imgY - 64) * scale;
                             Multiset<MapColor> colorCount = LinkedHashMultiset.create();
-                            LevelChunk chunk = level.getChunk(
+                            LevelChunk chunk = level.getChunkIfLoaded( // Paper - Maps shouldn't load chunks
                                 SectionPos.blockToSectionCoord(averagingAreaMinX), SectionPos.blockToSectionCoord(averagingAreaMinZ)
                             );
-                            if (!chunk.isEmpty()) {
+                            if (chunk != null && !chunk.isEmpty()) { // Paper - Maps shouldn't load chunks
                                 int waterDepth = 0;
                                 double averageAreaHeight = 0.0;
                                 if (level.dimensionType().hasCeiling()) {
@@ -200,6 +_,7 @@
     public static void renderBiomePreviewMap(ServerLevel level, ItemStack mapItemStack) {
         MapItemSavedData data = getSavedData(mapItemStack, level);
         if (data != null) {
+            data.isExplorerMap = true; // Purpur - Explorer Map API
             if (level.dimension() == data.dimension) {
                 int scale = 1 << data.scale;
                 int centerX = data.centerX;
