--- a/net/minecraft/world/item/MinecartItem.java
+++ b/net/minecraft/world/item/MinecartItem.java
@@ -30,12 +_,13 @@
         BlockPos pos = context.getClickedPos();
         BlockState blockState = level.getBlockState(pos);
         if (!blockState.is(BlockTags.RAILS)) {
-            return InteractionResult.FAIL;
+            if (!level.purpurConfig.minecartPlaceAnywhere) return InteractionResult.FAIL; // Purpur - Minecart settings and WASD controls
+            if (blockState.isSolid()) pos = pos.relative(context.getClickedFace());
         }
 
         ItemStack itemStack = context.getItemInHand();
         RailShape shape = blockState.getBlock() instanceof BaseRailBlock
-            ? blockState.getValue(((BaseRailBlock)blockState.getBlock()).getShapeProperty())
+            ? ((BaseRailBlock)blockState.getBlock()).getRailDirection(blockState, level, pos, null)
             : RailShape.NORTH_SOUTH;
         double offset = 0.0;
         if (shape.isSlope()) {
@@ -59,7 +_,13 @@
         }
 
         if (level instanceof ServerLevel serverLevel) {
-            serverLevel.addFreshEntity(cart);
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPlaceEvent(context, cart).isCancelled()) {
+                if (context.getPlayer() != null) context.getPlayer().containerMenu.forceHeldSlot(context.getHand()); // Paper - Fix inventory desync
+                return InteractionResult.FAIL;
+            }
+            // CraftBukkit end
+            if (!serverLevel.addFreshEntity(cart)) return InteractionResult.PASS; // CraftBukkit
             serverLevel.gameEvent(GameEvent.ENTITY_PLACE, pos, GameEvent.Context.of(context.getPlayer(), serverLevel.getBlockState(pos.below())));
         }
 
