--- a/net/minecraft/world/item/MobBucketItem.java
+++ b/net/minecraft/world/item/MobBucketItem.java
@@ -18,6 +_,7 @@
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.BlockHitResult;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 import org.jspecify.annotations.Nullable;
 
 public class MobBucketItem extends BucketItem {
@@ -52,6 +_,7 @@
         }
 
         if (mob != null) {
+            mob.spawnReason(CreatureSpawnEvent.SpawnReason.BUCKET); // CraftBukkit
             level.addFreshEntityWithPassengers(mob);
             mob.playAmbientSound();
         }
