--- a/net/minecraft/world/item/OminousBottleItem.java
+++ b/net/minecraft/world/item/OminousBottleItem.java
@@ -36,6 +_,7 @@
         if (!p_338693_.isClientSide) {
             p_338693_.playSound(null, p_338370_.blockPosition(), SoundEvents.OMINOUS_BOTTLE_DISPOSE, p_338370_.getSoundSource(), 1.0F, 1.0F);
             Integer integer = p_338871_.getOrDefault(DataComponents.OMINOUS_BOTTLE_AMPLIFIER, Integer.valueOf(0));
+            p_338370_.addEffectCause(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.POTION_DRINK); // Paper - Add missing effect cause
             p_338370_.addEffect(new MobEffectInstance(MobEffects.BAD_OMEN, 120000, integer, false, false, true));
         }
 
