--- a/net/minecraft/world/item/PotionItem.java
+++ b/net/minecraft/world/item/PotionItem.java
@@ -55,6 +_,7 @@
                 if (p_330883_.getEffect().value().isInstantenous()) {
                     p_330883_.getEffect().value().applyInstantenousEffect(player, player, p_42986_, p_330883_.getAmplifier(), 1.0);
                 } else {
+                    p_42986_.addEffectCause(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.POTION_DRINK); // CraftBukkit
                     p_42986_.addEffect(p_330883_);
                 }
             });
