--- a/net/minecraft/world/item/ProjectileItem.java
+++ b/net/minecraft/world/item/ProjectileItem.java
@@ -20,7 +_,7 @@
         projectile.shoot(xd, yd, zd, pow, uncertainty);
     }
 
-    record DispenseConfig(ProjectileItem.PositionFunction positionFunction, float uncertainty, float power, OptionalInt overrideDispenseEvent) {
+    public record DispenseConfig(ProjectileItem.PositionFunction positionFunction, float uncertainty, float power, OptionalInt overrideDispenseEvent) {
         public static final ProjectileItem.DispenseConfig DEFAULT = builder().build();
 
         public static ProjectileItem.DispenseConfig.Builder builder() {
@@ -62,7 +_,7 @@
     }
 
     @FunctionalInterface
-    interface PositionFunction {
+    public interface PositionFunction {
         Position getDispensePosition(final BlockSource source, final Direction direction);
     }
 }
