--- a/net/minecraft/world/item/SignItem.java
+++ b/net/minecraft/world/item/SignItem.java
@@ -11,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class SignItem extends StandingAndWallBlockItem {
+    public static BlockPos openSign; // CraftBukkit
     public SignItem(Item.Properties p_43126_, Block p_43127_, Block p_43128_) {
         super(p_43127_, p_43128_, p_43126_, Direction.DOWN);
     }
@@ -28,6 +_,7 @@
             && p_43131_.getBlockEntity(p_43130_) instanceof SignBlockEntity signblockentity
             && p_43131_.getBlockState(p_43130_).getBlock() instanceof SignBlock signblock) {
             signblock.openTextEdit(p_43132_, signblockentity, true);
+            SignItem.openSign = p_43130_;
         }
 
         return flag;
