--- /dev/null
+++ b/net/minecraft/world/item/SignItem.java
@@ -1,0 +_,39 @@
+package net.minecraft.world.item;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.SignBlock;
+import net.minecraft.world.level.block.entity.SignBlockEntity;
+import net.minecraft.world.level.block.state.BlockState;
+import org.jspecify.annotations.Nullable;
+
+public class SignItem extends StandingAndWallBlockItem {
+
+    public static BlockPos openSign; // CraftBukkit
+
+    public SignItem(Block sign, Block wallSign, Item.Properties properties) {
+        super(sign, wallSign, Direction.DOWN, properties);
+    }
+
+    public SignItem(Item.Properties properties, Block sign, Block wallSign, Direction direction) {
+        super(sign, wallSign, direction, properties);
+    }
+
+    @Override
+    protected boolean updateCustomBlockEntityTag(BlockPos pos, Level level, @Nullable Player player, ItemStack itemStack, BlockState placedState) {
+        boolean success = super.updateCustomBlockEntityTag(pos, level, player, itemStack, placedState);
+        if (!level.isClientSide()
+            && !success
+            && player != null
+            && level.getBlockEntity(pos) instanceof SignBlockEntity signEntity
+            && level.getBlockState(pos).getBlock() instanceof SignBlock sign) {
+            sign.openTextEdit(player, signEntity, true);
+            SignItem.openSign = pos;
+        }
+
+        return success;
+    }
+}
