--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -12,6 +_,7 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.permissions.Permissions;
 import net.minecraft.stats.Stats;
+import net.minecraft.world.Difficulty;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.AgeableMob;
@@ -164,6 +_,7 @@
 
         offspring.snapTo(pos.x(), pos.y(), pos.z(), 0.0F, 0.0F);
         offspring.applyComponentsFromItemStack(spawnEggStack);
+        offspring.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG); // CraftBukkit
         level.addFreshEntityWithPassengers(offspring);
         spawnEggStack.consume(1, player);
         return Optional.of(offspring);
@@ -179,5 +_,25 @@
         }
 
         return false;
+    }
+
+    /**
+     * {@return the dispense behavior to register or null to register the default}
+     */
+    protected net.minecraft.core.dispenser.@Nullable DispenseItemBehavior createDispenseBehavior() {
+        return null;
+    }
+
+    // TODO 26.1: It's questionable that this is even needed anymore
+    @net.neoforged.bus.api.SubscribeEvent(priority = net.neoforged.bus.api.EventPriority.LOWEST)
+    private static void registerDispenseBehavior(final net.neoforged.fml.event.lifecycle.FMLCommonSetupEvent event) {
+        event.enqueueWork(() -> BuiltInRegistries.ITEM.forEach(item -> {
+            if (item instanceof SpawnEggItem egg && !net.minecraft.world.level.block.DispenserBlock.DISPENSER_REGISTRY.containsKey(egg)) {
+                var beh = egg.createDispenseBehavior();
+                if (beh != null) {
+                    net.minecraft.world.level.block.DispenserBlock.registerBehavior(egg, beh);
+                }
+            }
+        }));
     }
 }
