--- a/net/minecraft/world/item/SplashPotionItem.java
+++ b/net/minecraft/world/item/SplashPotionItem.java
@@ -14,16 +_,23 @@
 
     @Override
     public InteractionResultHolder<ItemStack> use(Level p_43243_, Player p_43244_, InteractionHand p_43245_) {
-        p_43243_.playSound(
-            null,
-            p_43244_.getX(),
-            p_43244_.getY(),
-            p_43244_.getZ(),
-            SoundEvents.SPLASH_POTION_THROW,
-            SoundSource.PLAYERS,
-            0.5F,
-            0.4F / (p_43243_.getRandom().nextFloat() * 0.4F + 0.8F)
-        );
-        return super.use(p_43243_, p_43244_, p_43245_);
+        // Paper start - PlayerLaunchProjectileEvent
+        InteractionResultHolder<ItemStack> wrapper = super.use(p_43243_, p_43244_, p_43245_);
+        if (wrapper.getResult() != net.minecraft.world.InteractionResult.FAIL) {
+            // Paper end - PlayerLaunchProjectileEvent
+            p_43243_.playSound(
+                    null,
+                    p_43244_.getX(),
+                    p_43244_.getY(),
+                    p_43244_.getZ(),
+                    SoundEvents.SPLASH_POTION_THROW,
+                    SoundSource.PLAYERS,
+                    0.5F,
+                    0.4F / (p_43243_.getRandom().nextFloat() * 0.4F + 0.8F)
+            );
+            // Paper start - PlayerLaunchProjectileEvent
+        }
+        return wrapper;
+        // Paper end - PlayerLaunchProjectileEvent
     }
 }
