--- a/net/minecraft/world/item/StandingAndWallBlockItem.java
+++ b/net/minecraft/world/item/StandingAndWallBlockItem.java
@@ -11,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 
 public class StandingAndWallBlockItem extends BlockItem {
-    public final Block wallBlock;
+    protected final Block wallBlock;
     private final Direction attachmentDirection;
 
     public StandingAndWallBlockItem(Block p_248873_, Block p_251044_, Item.Properties p_249308_, Direction p_250800_) {
@@ -49,5 +_,10 @@
     public void registerBlocks(Map<Block, Item> p_43252_, Item p_43253_) {
         super.registerBlocks(p_43252_, p_43253_);
         p_43252_.put(this.wallBlock, p_43253_);
+    }
+
+    public void removeFromBlockToItemMap(Map<Block, Item> blockToItemMap, Item itemIn) {
+        super.removeFromBlockToItemMap(blockToItemMap, itemIn);
+        blockToItemMap.remove(this.wallBlock);
     }
 }
