--- a/net/minecraft/world/item/ThrowablePotionItem.java
+++ b/net/minecraft/world/item/ThrowablePotionItem.java
@@ -22,11 +_,29 @@
             ThrownPotion thrownpotion = new ThrownPotion(p_43303_, p_43304_);
             thrownpotion.setItem(itemstack);
             thrownpotion.shootFromRotation(p_43304_, p_43304_.getXRot(), p_43304_.getYRot(), -20.0F, 0.5F, 1.0F);
-            p_43303_.addFreshEntity(thrownpotion);
+            // Paper start - PlayerLaunchProjectileEvent
+            com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent event = new com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent((org.bukkit.entity.Player) p_43304_.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), (org.bukkit.entity.Projectile) thrownpotion.getBukkitEntity());
+            if (event.callEvent() && p_43303_.addFreshEntity(thrownpotion)) {
+                if (event.shouldConsume()) {
+                    itemstack.consume(1, p_43304_);
+                } else if (p_43304_ instanceof net.minecraft.server.level.ServerPlayer) {
+                    ((net.minecraft.server.level.ServerPlayer) p_43304_).getBukkitEntity().updateInventory();
+                }
+
+                p_43304_.awardStat(Stats.ITEM_USED.get(this));
+            } else {
+                if (p_43304_ instanceof net.minecraft.server.level.ServerPlayer) {
+                    ((net.minecraft.server.level.ServerPlayer) p_43304_).getBukkitEntity().updateInventory();
+                }
+                return InteractionResultHolder.fail(itemstack);
+            }
+            // Paper end - PlayerLaunchProjectileEvent
         }
 
+        /* // Paper start - PlayerLaunchProjectileEvent; moved up
         p_43304_.awardStat(Stats.ITEM_USED.get(this));
-        itemstack.consume(1, p_43304_);
+        itemStack.consume(1, p_43304_);
+        */ // Paper end
         return InteractionResultHolder.sidedSuccess(itemstack, p_43303_.isClientSide());
     }
 
