--- a/net/minecraft/world/item/TooltipFlag.java
+++ b/net/minecraft/world/item/TooltipFlag.java
@@ -1,14 +_,33 @@
 package net.minecraft.world.item;
 
-import net.neoforged.neoforge.common.extensions.TooltipFlagExtension;
-
-public interface TooltipFlag extends TooltipFlagExtension {
+public interface TooltipFlag {
     TooltipFlag.Default NORMAL = new TooltipFlag.Default(false, false);
     TooltipFlag.Default ADVANCED = new TooltipFlag.Default(true, false);
 
     boolean isAdvanced();
 
     boolean isCreative();
+
+    /**
+     * Neo: Returns the state of the Control key (as reported by Screen) on the client, or {@code false} on the server.
+     */
+    default boolean hasControlDown() {
+        return false;
+    }
+
+    /**
+     * Neo: Returns the state of the Shift key (as reported by Screen) on the client, or {@code false} on the server.
+     */
+    default boolean hasShiftDown() {
+        return false;
+    }
+
+    /**
+     * Neo: Returns the state of the Alt key (as reported by Screen) on the client, or {@code false} on the server.
+     */
+    default boolean hasAltDown() {
+        return false;
+    }
 
     record Default(boolean advanced, boolean creative) implements TooltipFlag {
         @Override
