--- a/net/minecraft/world/item/TridentItem.java
+++ b/net/minecraft/world/item/TridentItem.java
@@ -68,57 +_,89 @@
             int timeHeld = this.getUseDuration(itemStack, entity) - remainingTime;
             if (timeHeld < 10) {
                 return false;
-            }
-
-            float riptideStrength = EnchantmentHelper.getTridentSpinAttackStrength(itemStack, player);
-            if (!(riptideStrength > 0.0F) || player.isInWaterOrRain() && !player.isPassenger()) {
-                if (itemStack.nextDamageWillBreak()) {
-                    return false;
-                }
-
-                Holder<SoundEvent> sound = EnchantmentHelper.pickHighestLevel(itemStack, EnchantmentEffectComponents.TRIDENT_SOUND)
-                    .orElse(SoundEvents.TRIDENT_THROW);
-                player.awardStat(Stats.ITEM_USED.get(this));
-                if (level instanceof ServerLevel serverLevel) {
-                    itemStack.hurtWithoutBreaking(1, player);
-                    if (riptideStrength == 0.0F) {
-                        ItemStack thrownItemStack = itemStack.consumeAndReturn(1, player);
-                        ThrownTrident trident = Projectile.spawnProjectileFromRotation(
-                            ThrownTrident::new, serverLevel, thrownItemStack, player, 0.0F, 2.5F, 1.0F
-                        );
-                        if (player.hasInfiniteMaterials()) {
-                            trident.pickup = AbstractArrow.Pickup.CREATIVE_ONLY;
-                        }
-
-                        level.playSound(null, trident, sound.value(), SoundSource.PLAYERS, 1.0F, 1.0F);
-                        return true;
-                    }
-                }
-
-                if (riptideStrength > 0.0F) {
-                    float yRot = player.getYRot();
-                    float xRot = player.getXRot();
-                    float xd = -Mth.sin(yRot * (float) (Math.PI / 180.0)) * Mth.cos(xRot * (float) (Math.PI / 180.0));
-                    float yd = -Mth.sin(xRot * (float) (Math.PI / 180.0));
-                    float zd = Mth.cos(yRot * (float) (Math.PI / 180.0)) * Mth.cos(xRot * (float) (Math.PI / 180.0));
-                    float dist = Mth.sqrt(xd * xd + yd * yd + zd * zd);
-                    xd *= riptideStrength / dist;
-                    yd *= riptideStrength / dist;
-                    zd *= riptideStrength / dist;
-                    player.push(xd, yd, zd);
-                    player.startAutoSpinAttack(20, 8.0F, itemStack);
-                    if (player.onGround()) {
-                        float heightDifference = 1.1999999F;
-                        player.move(MoverType.SELF, new Vec3(0.0, 1.1999999F, 0.0));
-                    }
-
-                    level.playSound(null, player, sound.value(), SoundSource.PLAYERS, 1.0F, 1.0F);
-                    return true;
-                } else {
-                    return false;
-                }
             } else {
-                return false;
+                float riptideStrength = EnchantmentHelper.getTridentSpinAttackStrength(itemStack, player);
+                if (!(riptideStrength > 0.0F) || player.isInWaterOrRain() && !player.isPassenger()) {
+                    if (itemStack.nextDamageWillBreak()) {
+                        return false;
+                    } else {
+                        Holder<SoundEvent> sound = EnchantmentHelper.pickHighestLevel(itemStack, EnchantmentEffectComponents.TRIDENT_SOUND)
+                            .orElse(SoundEvents.TRIDENT_THROW);
+                        player.awardStat(Stats.ITEM_USED.get(this));
+                        if (level instanceof ServerLevel serverLevel) {
+                            // itemStack.hurtWithoutBreaking(1, player); // CraftBukkit - moved down
+                            if (riptideStrength == 0.0F) {
+                                ItemStack thrownItemStack = itemStack.copyWithCount(1); // Paper
+                                Projectile.Delayed<ThrownTrident> tridentDelayed = Projectile.spawnProjectileFromRotationDelayed( // Paper - PlayerLaunchProjectileEvent(
+                                        ThrownTrident::new, serverLevel, thrownItemStack, player, 0.0F, 2.5F, (float) serverLevel.purpurConfig.tridentProjectileOffset  // Purpur - Projectile offset config
+                                );
+                                // Paper start - PlayerLaunchProjectileEvent
+                                com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent event = new com.destroystokyo.paper.event.player.PlayerLaunchProjectileEvent((org.bukkit.entity.Player) player.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemStack), (org.bukkit.entity.Projectile) tridentDelayed.projectile().getBukkitEntity());
+                                if (!event.callEvent() || !tridentDelayed.attemptSpawn()) {
+                                    // CraftBukkit start
+                                    // Paper end - PlayerLaunchProjectileEvent
+                                    return false;
+                                }
+                                ThrownTrident trident = tridentDelayed.projectile(); // Paper - PlayerLaunchProjectileEvent
+                                trident.setActualEnchantments(itemStack.getEnchantments()); // Purpur - Add an option to fix MC-3304 projectile looting
+                                if (event.shouldConsume()) {
+                                    thrownItemStack.hurtWithoutBreaking(1, player); // Paper - PlayerLaunchProjectileEvent - use thrownItemStack; pickup item damage
+                                }
+                                trident.setPickupItemStack(thrownItemStack.copy()); // SPIGOT-4511 update since damage call moved - use thrownItemStack; count = 1
+                                if (event.shouldConsume()) {
+                                    itemStack.consume(1, player);
+                                }
+                                // CraftBukkit end
+                                if (player.hasInfiniteMaterials()) {
+                                    trident.pickup = AbstractArrow.Pickup.CREATIVE_ONLY;
+                                }
+
+                                level.playSound(null, trident, sound.value(), SoundSource.PLAYERS, 1.0F, 1.0F);
+                                return true;
+                                // CraftBukkit start - SPIGOT-5458 also need in this branch :(
+                            } else {
+                                itemStack.hurtWithoutBreaking(1, player);
+                                // CraftBukkit end
+                            }
+                        }
+
+                        if (riptideStrength > 0.0F) {
+                            float yRot = player.getYRot();
+                            float xRot = player.getXRot();
+                            float xd = -Mth.sin(yRot * (float) (Math.PI / 180.0)) * Mth.cos(xRot * (float) (Math.PI / 180.0));
+                            float yd = -Mth.sin(xRot * (float) (Math.PI / 180.0));
+                            float zd = Mth.cos(yRot * (float) (Math.PI / 180.0)) * Mth.cos(xRot * (float) (Math.PI / 180.0));
+                            float dist = Mth.sqrt(xd * xd + yd * yd + zd * zd);
+                            xd *= riptideStrength / dist;
+                            yd *= riptideStrength / dist;
+                            zd *= riptideStrength / dist;
+                            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerRiptideEvent(player, itemStack, xd, yd, zd)) return false; // Paper - Add player riptide event
+                            // Purpur start - Implement elytra settings
+                            List<net.minecraft.world.entity.EquipmentSlot> list = net.minecraft.world.entity.EquipmentSlot.VALUES.stream().filter((enumitemslot) -> LivingEntity.canGlideUsing(entity.getItemBySlot(enumitemslot), enumitemslot)).toList();
+                            if (!list.isEmpty()) {
+                                net.minecraft.world.entity.EquipmentSlot enumitemslot = net.minecraft.util.Util.getRandom(list, entity.random);
+                                ItemStack glideItem = entity.getItemBySlot(enumitemslot);
+                                if (glideItem.has(net.minecraft.core.component.DataComponents.GLIDER) && level.purpurConfig.elytraDamagePerTridentBoost > 0) {
+                                    glideItem.hurtAndBreak(level.purpurConfig.elytraDamagePerTridentBoost, entity, enumitemslot);
+                                }
+                            }
+                            // Purpur end - Implement elytra settings
+                            player.push(xd, yd, zd);
+                            player.startAutoSpinAttack(20, 8.0F, itemStack);
+                            if (player.onGround()) {
+                                float heightDifference = 1.1999999F;
+                                player.move(MoverType.SELF, new Vec3(0.0, 1.1999999F, 0.0));
+                            }
+
+                            level.playSound(null, player, sound.value(), SoundSource.PLAYERS, 1.0F, 1.0F);
+                            return true;
+                        } else {
+                            return false;
+                        }
+                    }
+                } else {
+                    return false;
+                }
             }
         } else {
             return false;
@@ -145,5 +_,10 @@
         ThrownTrident trident = new ThrownTrident(level, position.x(), position.y(), position.z(), itemStack.copyWithCount(1));
         trident.pickup = AbstractArrow.Pickup.ALLOWED;
         return trident;
+    }
+
+    @Override
+    public boolean canPerformAction(ItemInstance stack, net.neoforged.neoforge.common.ItemAbility itemAbility) {
+        return net.neoforged.neoforge.common.ItemAbilities.DEFAULT_TRIDENT_ACTIONS.contains(itemAbility);
     }
 }
