--- a/net/minecraft/world/item/WrittenBookItem.java
+++ b/net/minecraft/world/item/WrittenBookItem.java
@@ -54,7 +_,7 @@
 
     public static boolean resolveBookComponents(ItemStack p_43462_, CommandSourceStack p_43463_, @Nullable Player p_43464_) {
         WrittenBookContent writtenbookcontent = p_43462_.get(DataComponents.WRITTEN_BOOK_CONTENT);
-        if (writtenbookcontent != null && !writtenbookcontent.resolved()) {
+        if (io.papermc.paper.configuration.GlobalConfiguration.get().itemValidation.resolveSelectorsInBooks && writtenbookcontent != null && !writtenbookcontent.resolved()) { // Paper - Disable component selector resolving in books by default
             WrittenBookContent writtenbookcontent1 = writtenbookcontent.resolve(p_43463_, p_43464_);
             if (writtenbookcontent1 != null) {
                 p_43462_.set(DataComponents.WRITTEN_BOOK_CONTENT, writtenbookcontent1);
