--- a/com/mojang/realmsclient/util/JsonUtils.java
+++ b/com/mojang/realmsclient/util/JsonUtils.java
@@ -91,10 +_,4 @@
         JsonElement element = node.get(key);
         return element != null ? Instant.ofEpochMilli(Long.parseLong(element.getAsString())) : Instant.EPOCH;
     }
-
-    @Contract("_,_,!null->!null;_,_,null->_")
-    public static @Nullable Instant getEpochMillisOr(String key, JsonObject node, @Nullable Instant defaultValue) {
-        JsonElement element = node.get(key);
-        return element != null && !element.isJsonNull() ? Instant.ofEpochMilli(Long.parseLong(element.getAsString())) : defaultValue;
-    }
 }
