--- a/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/AbstractSkeleton.java
@@ -138,8 +_,8 @@
         if (this.level() != null && !this.level().isClientSide()) {
             this.goalSelector.removeGoal(this.meleeGoal);
             this.goalSelector.removeGoal(this.bowGoal);
-            ItemStack usedWeapon = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, Items.BOW));
-            if (usedWeapon.is(Items.BOW)) {
+            ItemStack itemstack = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.world.item.BowItem));
+            if (itemstack.getItem() instanceof net.minecraft.world.item.BowItem) {
                 int minAttackInterval = this.getHardAttackInterval();
                 if (this.level().getDifficulty() != Difficulty.HARD) {
                     minAttackInterval = this.getAttackInterval();
@@ -163,17 +_,29 @@
 
     @Override
     public void performRangedAttack(LivingEntity target, float power) {
-        ItemStack bowItem = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, Items.BOW));
+        ItemStack bowItem = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.world.item.BowItem));
         ItemStack projectile = this.getProjectile(bowItem);
         AbstractArrow arrow = this.getArrow(projectile, power, bowItem);
+        if (bowItem.getItem() instanceof net.minecraft.world.item.ProjectileWeaponItem weaponItem)
+            arrow = weaponItem.customArrow(arrow, projectile, bowItem);
         double xd = target.getX() - this.getX();
         double yd = target.getY(0.3333333333333333) - arrow.getY();
         double zd = target.getZ() - this.getZ();
         double distanceToTarget = Math.sqrt(xd * xd + zd * zd);
         if (this.level() instanceof ServerLevel serverLevel) {
-            Projectile.spawnProjectileUsingShoot(
-                arrow, serverLevel, projectile, xd, yd + distanceToTarget * 0.2F, zd, 1.6F, 14 - serverLevel.getDifficulty().getId() * 4
-            );
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(this, this.getMainHandItem(), null, arrow, net.minecraft.world.InteractionHand.MAIN_HAND, 0.8F, true);
+            if (event.isCancelled()) {
+                event.getProjectile().remove();
+                return;
+            }
+
+            if (event.getProjectile() == arrow.getBukkitEntity()) {
+                Projectile.spawnProjectileUsingShoot(
+                    arrow, serverLevel, projectile, xd, yd + distanceToTarget * 0.2F, zd, 1.6F, 14 - serverLevel.getDifficulty().getId() * 4
+                );
+            }
+            // CraftBukkit end
         }
 
         this.playSound(SoundEvents.SKELETON_SHOOT, 1.0F, 1.0F / (this.getRandom().nextFloat() * 0.4F + 0.8F));
@@ -185,7 +_,7 @@
 
     @Override
     public boolean canUseNonMeleeWeapon(ItemStack item) {
-        return item.getItem() == Items.BOW;
+        return item.is(net.neoforged.neoforge.common.Tags.Items.SKELETON_USABLE_BOWS);
     }
 
     @Override
@@ -200,8 +_,10 @@
     }
 
     @Override
-    public void onEquipItem(EquipmentSlot slot, ItemStack oldStack, ItemStack stack) {
-        super.onEquipItem(slot, oldStack, stack);
+    // CraftBukkit start
+    public void onEquipItem(EquipmentSlot slot, ItemStack oldStack, ItemStack stack, boolean silent) {
+        super.onEquipItem(slot, oldStack, stack, silent);
+        // CraftBukkit end
         if (!this.level().isClientSide()) {
             this.reassessWeaponGoal();
         }
