--- a/net/minecraft/world/entity/monster/skeleton/Bogged.java
+++ b/net/minecraft/world/entity/monster/skeleton/Bogged.java
@@ -70,9 +_,15 @@
     @Override
     protected InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemStack = player.getItemInHand(hand);
-        if (itemStack.is(Items.SHEARS) && this.readyForShearing()) {
-            if (this.level() instanceof ServerLevel level) {
-                this.shear(level, SoundSource.PLAYERS, itemStack);
+        if (false && itemStack.is(Items.SHEARS) && this.readyForShearing()) { // Neo: Shear logic is handled by IShearable
+            if (this.level() instanceof ServerLevel serverLevel) {
+                // CraftBukkit start
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerShearEntityEvent(player, this, itemStack, hand)) {
+                    this.getEntityData().markDirty(DATA_SHEARED); // CraftBukkit - mark dirty to restore sheared state to clients
+                    return InteractionResult.PASS;
+                }
+                // CraftBukkit end
+                this.shear(serverLevel, SoundSource.PLAYERS, itemStack);
                 this.gameEvent(GameEvent.SHEAR, player);
                 itemStack.hurtAndBreak(1, player, hand.asEquipmentSlot());
             }
