--- a/net/minecraft/world/entity/monster/skeleton/Parched.java
+++ b/net/minecraft/world/entity/monster/skeleton/Parched.java
@@ -17,6 +_,7 @@
 public class Parched extends AbstractSkeleton {
     public Parched(EntityType<? extends AbstractSkeleton> type, Level level) {
         super(type, level);
+        this.setShouldBurnInDay(false); // Purpur - API for any mob to burn daylight
     }
 
     @Override
