--- a/net/minecraft/world/entity/monster/skeleton/Skeleton.java
+++ b/net/minecraft/world/entity/monster/skeleton/Skeleton.java
@@ -93,11 +_,18 @@
     }
 
     protected void doFreezeConversion() {
-        this.convertTo(EntityTypes.STRAY, ConversionParams.single(this, true, true), stray -> {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityTypes.STRAY, (timer) -> this.conversionTime = timer)) return;
+        Stray converted = this.convertTo(EntityTypes.STRAY, ConversionParams.single(this, true, true), stray -> { // CraftBukkit
+            net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, stray);
             if (!this.isSilent()) {
                 this.level().levelEvent(null, 1048, this.blockPosition(), 0);
             }
-        });
+        // CraftBukkit start
+        }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.FROZEN, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.FROZEN);
+        if (converted == null) {
+            ((org.bukkit.entity.Skeleton) getBukkitEntity()).setConversionTime(-1); // CraftBukkit - SPIGOT-7997: End conversion to stop event spam
+        }
+        // CraftBukkit end
     }
 
     @Override
