--- a/net/minecraft/world/entity/animal/dolphin/Dolphin.java
+++ b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
@@ -64,6 +_,9 @@
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
+import org.bukkit.event.entity.EntityRemoveEvent;
 import org.jspecify.annotations.Nullable;
 
 public class Dolphin extends AgeableWaterCreature {
@@ -84,6 +_,13 @@
         this.setCanPickUpLoot(true);
     }
 
+    // CraftBukkit start - SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
+    @Override
+    public int getDefaultMaxAirSupply() {
+        return TOTAL_AIR_SUPPLY;
+    }
+    // CraftBukkit end
+
     @Override
     public @Nullable SpawnGroupData finalizeSpawn(
         ServerLevelAccessor level, DifficultyInstance difficulty, EntitySpawnReason spawnReason, @Nullable SpawnGroupData groupData
@@ -183,7 +_,7 @@
 
     @Override
     public int getMaxAirSupply() {
-        return 4800;
+        return maxAirTicks; // CraftBukkit - SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
     }
 
     @Override
@@ -216,10 +_,17 @@
         if (this.getItemBySlot(EquipmentSlot.MAINHAND).isEmpty()) {
             ItemStack itemStack = entity.getItem();
             if (this.canHoldItem(itemStack)) {
+                // CraftBukkit start - call EntityPickupItemEvent
+                if (CraftEventFactory.callEntityPickupItemEvent(this, entity, 0, false).isCancelled()) {
+                    return;
+                }
+                itemStack = entity.getItem(); // CraftBukkit- update ItemStack from event
+                // CraftBukkit start
                 this.onItemPickup(entity);
                 this.setItemSlot(EquipmentSlot.MAINHAND, itemStack);
                 this.setGuaranteedDrop(EquipmentSlot.MAINHAND);
                 this.take(entity, itemStack.getCount());
+                entity.removeReason(EntityRemoveEvent.Cause.PICKUP); // CraftBukkit - add Bukkit remove cause
                 entity.discard();
             }
         }
@@ -479,6 +_,7 @@
 
         @Override
         public void start() {
+            this.player.addEffectCause(EntityPotionEffectEvent.Cause.DOLPHIN); // CraftBukkit
             this.player.addEffect(new MobEffectInstance(MobEffects.DOLPHINS_GRACE, 100), this.dolphin);
         }
 
@@ -498,6 +_,7 @@
             }
 
             if (this.player.isSwimming() && this.player.level().getRandom().nextInt(6) == 0) {
+                this.player.addEffectCause(EntityPotionEffectEvent.Cause.DOLPHIN); // CraftBukkit
                 this.player.addEffect(new MobEffectInstance(MobEffects.DOLPHINS_GRACE, 100), this.dolphin);
             }
         }
