--- a/net/minecraft/client/resources/sounds/RidingEntitySoundInstance.java
+++ b/net/minecraft/client/resources/sounds/RidingEntitySoundInstance.java
@@ -30,9 +_,6 @@
         super(soundEvent, soundSource, SoundInstance.createUnseededRandom());
         this.player = player;
         this.entity = entity;
-        this.x = entity.getX();
-        this.y = entity.getY();
-        this.z = entity.getZ();
         this.underwaterSound = underwaterSound;
         this.volumeMin = volumeMin;
         this.volumeMax = volumeMax;
@@ -78,10 +_,6 @@
             } else {
                 this.volume = this.volumeMin;
             }
-
-            this.x = this.entity.getX();
-            this.y = this.entity.getY();
-            this.z = this.entity.getZ();
         }
     }
 }
