--- a/net/minecraft/client/resources/sounds/Sound.java
+++ b/net/minecraft/client/resources/sounds/Sound.java
@@ -83,10 +_,6 @@
         return this.attenuationDistance;
     }
 
-    public float getAttenuationDistance(float volume) {
-        return Math.max(volume, 1.0F) * this.attenuationDistance;
-    }
-
     @Override
     public String toString() {
         return "Sound[" + this.location + "]";
