--- a/net/minecraft/client/resources/sounds/SoundInstance.java
+++ b/net/minecraft/client/resources/sounds/SoundInstance.java
@@ -13,11 +_,9 @@
 public interface SoundInstance {
     Identifier getIdentifier();
 
-    @Nullable WeighedSoundEvents getOrResolve(SoundManager soundManager);
+    @Nullable WeighedSoundEvents resolve(SoundManager soundManager);
 
     @Nullable Sound getSound();
-
-    @Nullable WeighedSoundEvents getSoundEvent();
 
     SoundSource getSource();
 
