--- a/net/minecraft/client/renderer/texture/SpriteContents.java
+++ b/net/minecraft/client/renderer/texture/SpriteContents.java
@@ -382,11 +_,11 @@
             if (this.animationInfo.interpolateFrames) {
                 int newFrame = frames.get((this.frame + 1) % frames.size()).index;
                 renderPass.setPipeline(RenderSystem.getCompiledPipeline(RenderPipelines.ANIMATE_SPRITE_INTERPOLATE));
-                renderPass.setUniform("CurrentSprite", this.frameTexturesByIndex.get(oldFrame), sampler);
-                renderPass.setUniform("NextSprite", this.frameTexturesByIndex.get(newFrame), sampler);
+                renderPass.bindTexture("CurrentSprite", this.frameTexturesByIndex.get(oldFrame), sampler);
+                renderPass.bindTexture("NextSprite", this.frameTexturesByIndex.get(newFrame), sampler);
             } else if (this.isDirty) {
                 renderPass.setPipeline(RenderSystem.getCompiledPipeline(RenderPipelines.ANIMATE_SPRITE_BLIT));
-                renderPass.setUniform("Sprite", this.frameTexturesByIndex.get(oldFrame), sampler);
+                renderPass.bindTexture("Sprite", this.frameTexturesByIndex.get(oldFrame), sampler);
             }
 
             renderPass.setUniform("SpriteAnimationInfo", ubo);
