--- a/net/minecraft/client/renderer/texture/SpriteContents.java
+++ b/net/minecraft/client/renderer/texture/SpriteContents.java
@@ -51,8 +_,8 @@
     private final int width;
     private final int height;
     private final NativeImage originalImage;
-    public NativeImage[] byMipLevel;
-    final SpriteContents.@Nullable AnimatedTexture animatedTexture;
+    private NativeImage[] byMipLevel;
+    private final SpriteContents.@Nullable AnimatedTexture animatedTexture;
     private final List<MetadataSectionType.WithValue<?>> additionalMetadata;
     private final MipmapStrategy mipmapStrategy;
     private final float alphaCutoffBias;
@@ -259,7 +_,7 @@
         RenderSystem.getDevice().createCommandEncoder().writeToTexture(destination, this.byMipLevel[level], level, 0, 0, 0);
     }
 
-    class AnimatedTexture {
+    private class AnimatedTexture {
         private final List<SpriteContents.FrameInfo> frames;
         private final IntList uniqueFrames;
         private final int frameRowSize;
@@ -272,11 +_,11 @@
             this.uniqueFrames = IntArrayList.toList(frames.stream().mapToInt(SpriteContents.FrameInfo::index).distinct());
         }
 
-        int getFrameX(int index) {
+        private int getFrameX(int index) {
             return index % this.frameRowSize;
         }
 
-        int getFrameY(int index) {
+        private int getFrameY(int index) {
             return index / this.frameRowSize;
         }
 
