--- a/net/minecraft/client/renderer/texture/SpriteLoader.java
+++ b/net/minecraft/client/renderer/texture/SpriteLoader.java
@@ -72,7 +_,7 @@
             int minSize = Math.min(minTexelSize, lowestOneBit);
             int minPowerOfTwo = Mth.log2(minSize);
             int mipLevel;
-            if (minPowerOfTwo < maxMipmapLevels) {
+            if (false) { // Forge: Do not lower the mipmap level
                 LOGGER.warn("{}: dropping miplevel from {} to {}, because of minimum power of two: {}", this.location, maxMipmapLevels, minPowerOfTwo, minSize);
                 mipLevel = minPowerOfTwo;
             } else {
@@ -129,7 +_,7 @@
         ResourceManager manager, Identifier atlasInfoLocation, int maxMipmapLevels, Executor taskExecutor, Set<MetadataSectionType<?>> additionalMetadata
     ) {
         SpriteResourceLoader spriteResourceLoader = SpriteResourceLoader.create(additionalMetadata);
-        return CompletableFuture.<List<SpriteSource.Loader>>supplyAsync(() -> SpriteSourceList.load(manager, atlasInfoLocation).list(manager), taskExecutor)
+        return CompletableFuture.<List<SpriteSource.Loader>>supplyAsync(() -> SpriteSourceList.load(manager, atlasInfoLocation).list(manager, additionalMetadata), taskExecutor)
             .thenCompose(sprites -> runSpriteSuppliers(spriteResourceLoader, (List<SpriteSource.Loader>)sprites, taskExecutor))
             .thenApply(resources -> this.stitch((List<SpriteContents>)resources, maxMipmapLevels, taskExecutor));
     }
