--- a/net/minecraft/client/renderer/texture/TextureAtlas.java
+++ b/net/minecraft/client/renderer/texture/TextureAtlas.java
@@ -187,7 +_,7 @@
                     renderPass.setPipeline(RenderSystem.getCompiledPipeline(RenderPipelines.ANIMATE_SPRITE_BLIT));
 
                     for (int i = 0; i < staticSprites.size(); i++) {
-                        renderPass.setUniform("Sprite", scratchTextures.get(i)[level], sampler);
+                        renderPass.bindTexture("Sprite", scratchTextures.get(i)[level], sampler);
                         renderPass.setUniform("SpriteAnimationInfo", ubo.slice(i * uboBlockSize + level * spriteUboSize, SpriteContents.UBO_SIZE));
                         renderPass.draw(6, 1, 0, 0);
                     }
