--- a/com/mojang/blaze3d/TracyFrameCapture.java
+++ b/com/mojang/blaze3d/TracyFrameCapture.java
@@ -86,7 +_,7 @@
                     .createRenderPass(() -> "Tracy blit", this.frameBufferView, Optional.empty())) {
                 RenderSystem.bindDefaultUniforms(renderPass);
                 renderPass.setPipeline(RenderSystem.getCompiledPipeline(RenderPipelines.TRACY_BLIT));
-                renderPass.setUniform("InSampler", captureTarget.getColorTextureView(), RenderSystem.getSamplerCache().getClampToEdge(FilterMode.LINEAR));
+                renderPass.bindTexture("InSampler", captureTarget.getColorTextureView(), RenderSystem.getSamplerCache().getClampToEdge(FilterMode.LINEAR));
                 renderPass.draw(3, 1, 0, 0);
             }
 
