--- a/net/minecraft/client/renderer/block/dispatch/multipart/MultiPartModel.java
+++ b/net/minecraft/client/renderer/block/dispatch/multipart/MultiPartModel.java
@@ -21,10 +_,8 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.model.DynamicBlockStateModel;
-
 @OnlyIn(Dist.CLIENT)
-public class MultiPartModel implements BlockStateModel, DynamicBlockStateModel {
+public class MultiPartModel implements BlockStateModel {
     private final MultiPartModel.SharedBakedState shared;
     private final BlockState blockState;
     private @Nullable List<BlockStateModel> models;
