--- a/net/minecraft/commands/arguments/DimensionArgument.java
+++ b/net/minecraft/commands/arguments/DimensionArgument.java
@@ -24,7 +_,7 @@
     private static final Collection<String> EXAMPLES = Stream.of(Level.OVERWORLD, Level.NETHER)
         .map(p_88814_ -> p_88814_.location().toString())
         .collect(Collectors.toList());
-    private static final DynamicCommandExceptionType ERROR_INVALID_VALUE = new DynamicCommandExceptionType(
+    public static final DynamicCommandExceptionType ERROR_INVALID_VALUE = new DynamicCommandExceptionType(
         p_304084_ -> Component.translatableEscape("argument.dimension.invalid", p_304084_)
     );
 
