--- a/net/minecraft/commands/arguments/EntityAnchorArgument.java
+++ b/net/minecraft/commands/arguments/EntityAnchorArgument.java
@@ -10,7 +_,6 @@
 import com.mojang.brigadier.suggestion.SuggestionsBuilder;
 import java.util.Arrays;
 import java.util.Collection;
-import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.CompletableFuture;
 import java.util.function.BiFunction;
