--- a/net/minecraft/commands/arguments/EntityArgument.java
+++ b/net/minecraft/commands/arguments/EntityArgument.java
@@ -94,6 +_,10 @@
         }
     }
 
+    public EntitySelector parse(StringReader stringreader, boolean overridePermissions) throws CommandSyntaxException {
+        return parse(stringreader, overridePermissions);
+    }
+
     public EntitySelector parse(StringReader p_91451_) throws CommandSyntaxException {
         int i = 0;
         EntitySelectorParser entityselectorparser = new EntitySelectorParser(p_91451_);
@@ -119,7 +_,7 @@
         if (p_91482_.getSource() instanceof SharedSuggestionProvider sharedsuggestionprovider) {
             StringReader stringreader = new StringReader(p_91483_.getInput());
             stringreader.setCursor(p_91483_.getStart());
-            EntitySelectorParser entityselectorparser = new EntitySelectorParser(stringreader, sharedsuggestionprovider.hasPermission(2));
+            EntitySelectorParser entityselectorparser = new EntitySelectorParser(stringreader, net.neoforged.neoforge.common.CommonHooks.canUseEntitySelectors(sharedsuggestionprovider));
 
             try {
                 entityselectorparser.parse();
