--- a/net/minecraft/commands/arguments/ResourceOrIdArgument.java
+++ b/net/minecraft/commands/arguments/ResourceOrIdArgument.java
@@ -40,10 +_,6 @@
 import net.minecraft.world.level.storage.loot.functions.LootItemFunction;
 import net.minecraft.world.level.storage.loot.functions.LootItemFunctions;
 import net.minecraft.world.level.storage.loot.predicates.LootItemCondition;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProvider;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProviders;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProvider;
-import net.minecraft.world.level.storage.loot.providers.number.ints.ContextIntProviders;
 import org.jspecify.annotations.Nullable;
 
 public class ResourceOrIdArgument<T> implements ArgumentType<Holder<T>> {
@@ -115,22 +_,6 @@
         return getResource(context, name);
     }
 
-    public static ResourceOrIdArgument.ContextFloatProviderArgument floatProvider(CommandBuildContext context) {
-        return new ResourceOrIdArgument.ContextFloatProviderArgument(context);
-    }
-
-    public static Holder<ContextFloatProvider> getFloatProvider(CommandContext<CommandSourceStack> context, String name) {
-        return getResource(context, name);
-    }
-
-    public static ResourceOrIdArgument.ContextIntProviderArgument intProvider(CommandBuildContext context) {
-        return new ResourceOrIdArgument.ContextIntProviderArgument(context);
-    }
-
-    public static Holder<ContextIntProvider> getIntProvider(CommandContext<CommandSourceStack> context, String name) {
-        return getResource(context, name);
-    }
-
     public static ResourceOrIdArgument.FeatureArgument feature(CommandBuildContext context) {
         return new ResourceOrIdArgument.FeatureArgument(context);
     }
@@ -170,18 +_,6 @@
     @Override
     public Collection<String> getExamples() {
         return EXAMPLES;
-    }
-
-    public static class ContextFloatProviderArgument extends ResourceOrIdArgument<ContextFloatProvider> {
-        protected ContextFloatProviderArgument(CommandBuildContext context) {
-            super(context, Registries.CONTEXT_FLOAT_PROVIDER, ContextFloatProviders.DIRECT_CODEC);
-        }
-    }
-
-    public static class ContextIntProviderArgument extends ResourceOrIdArgument<ContextIntProvider> {
-        protected ContextIntProviderArgument(CommandBuildContext context) {
-            super(context, Registries.CONTEXT_INT_PROVIDER, ContextIntProviders.DIRECT_CODEC);
-        }
     }
 
     public static class DialogArgument extends ResourceOrIdArgument<Dialog> {
