--- a/net/minecraft/data/worldgen/material/EndMaterialRules.java
+++ b/net/minecraft/data/worldgen/material/EndMaterialRules.java
@@ -5,17 +_,16 @@
 import net.minecraft.resources.Identifier;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.world.level.block.Blocks;
-import net.minecraft.world.level.levelgen.material.MaterialRules;
-import net.minecraft.world.level.levelgen.material.rule.MaterialRule;
+import net.minecraft.world.level.levelgen.SurfaceRules;
 
 public class EndMaterialRules {
-    public static final ResourceKey<MaterialRule> END = createKey("end");
+    public static final ResourceKey<SurfaceRules.RuleSource> END = createKey("end");
 
-    private static ResourceKey<MaterialRule> createKey(String name) {
+    private static ResourceKey<SurfaceRules.RuleSource> createKey(String name) {
         return ResourceKey.create(Registries.MATERIAL_RULE, Identifier.withDefaultNamespace(name));
     }
 
-    public static void bootstrap(BootstrapContext<MaterialRule> context) {
-        context.register(END, MaterialRules.state(Blocks.END_STONE.defaultBlockState()));
+    public static void bootstrap(BootstrapContext<SurfaceRules.RuleSource> context) {
+        context.register(END, SurfaceRules.state(Blocks.END_STONE.defaultBlockState()));
     }
 }
