--- a/net/minecraft/data/worldgen/material/OverworldMaterialRules.java
+++ b/net/minecraft/data/worldgen/material/OverworldMaterialRules.java
@@ -2,7 +_,6 @@
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableList.Builder;
-import java.util.List;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.registries.Registries;
 import net.minecraft.data.worldgen.BootstrapContext;
@@ -12,425 +_,413 @@
 import net.minecraft.world.level.biome.Biomes;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
-import net.minecraft.world.level.levelgen.NoiseRouterData;
 import net.minecraft.world.level.levelgen.Noises;
+import net.minecraft.world.level.levelgen.SurfaceRules;
 import net.minecraft.world.level.levelgen.VerticalAnchor;
-import net.minecraft.world.level.levelgen.densityfunction.DensityFunction;
-import net.minecraft.world.level.levelgen.material.MaterialRules;
-import net.minecraft.world.level.levelgen.material.condition.MaterialCondition;
-import net.minecraft.world.level.levelgen.material.rule.MaterialRule;
-import net.minecraft.world.level.levelgen.material.rule.OreVeinRule;
 
 public class OverworldMaterialRules {
-    public static final ResourceKey<MaterialRule> OVERWORLD = createKey("overworld");
-    public static final ResourceKey<MaterialRule> OVERWORLD_CAVES = createKey("overworld_caves");
-    public static final ResourceKey<MaterialRule> OVERWORLD_FLOATING_ISLANDS = createKey("overworld_floating_islands");
-    private static final ResourceKey<MaterialRule> SURFACE = createKey("overworld/surface");
-    private static final ResourceKey<MaterialRule> UNDERGROUND = createKey("overworld/underground");
-    private static final ResourceKey<MaterialRule> SULFUR_CAVE_BANDS = createKey("overworld/sulfur_cave_bands");
-    private static final ResourceKey<MaterialRule> COPPER_ORE_VEIN = createKey("overworld/copper_ore_vein");
-    private static final ResourceKey<MaterialRule> IRON_ORE_VEIN = createKey("overworld/iron_ore_vein");
-    private static final ResourceKey<MaterialRule> BIOME_SURFACE = createKey("overworld/biome_surface");
-    private static final ResourceKey<MaterialRule> DAPPLED_FOREST_SURFACE = createKey("overworld/biome_surface/dappled_forest");
-    private static final ResourceKey<MaterialRule> DEFAULT_BIOME_SURFACE = createKey("overworld/biome_surface/default");
-    private static final ResourceKey<MaterialRule> DRIPSTONE_CAVES_SURFACE = createKey("overworld/biome_surface/dripstone_caves");
-    private static final ResourceKey<MaterialRule> FROZEN_PEAKS_SURFACE = createKey("overworld/biome_surface/frozen_peaks");
-    private static final ResourceKey<MaterialRule> GROVE_SURFACE = createKey("overworld/biome_surface/grove");
-    private static final ResourceKey<MaterialRule> ICE_SPIKES_SURFACE = createKey("overworld/biome_surface/ice_spikes");
-    private static final ResourceKey<MaterialRule> JAGGED_PEAKS_SURFACE = createKey("overworld/biome_surface/jagged_peaks");
-    private static final ResourceKey<MaterialRule> MANGROVE_SWAMP_SURFACE = createKey("overworld/biome_surface/mangrove_swamp");
-    private static final ResourceKey<MaterialRule> MUSHROOM_FIELDS_SURFACE = createKey("overworld/biome_surface/mushroom_fields");
-    private static final ResourceKey<MaterialRule> OLD_GROWTH_PINE_TAIGA_SURFACE = createKey("overworld/biome_surface/old_growth_pine_taiga");
-    private static final ResourceKey<MaterialRule> SNOWY_SLOPES_SURFACE = createKey("overworld/biome_surface/snowy_slopes");
-    private static final ResourceKey<MaterialRule> STONY_PEAKS_SURFACE = createKey("overworld/biome_surface/stony_peaks");
-    private static final ResourceKey<MaterialRule> STONY_SHORE_SURFACE = createKey("overworld/biome_surface/stony_shore");
-    private static final ResourceKey<MaterialRule> SULFUR_CAVES_SURFACE = createKey("overworld/biome_surface/sulfur_caves");
-    private static final ResourceKey<MaterialRule> WINDSWEPT_GRAVELLY_HILLS_SURFACE = createKey("overworld/biome_surface/windswept_gravelly_hills");
-    private static final ResourceKey<MaterialRule> WINDSWEPT_HILLS_SURFACE = createKey("overworld/biome_surface/windswept_hills");
-    private static final ResourceKey<MaterialRule> WINDSWEPT_SAVANNA_SURFACE = createKey("overworld/biome_surface/windswept_savanna");
-    private static final ResourceKey<MaterialRule> UNDER_BIOME_SURFACE = createKey("overworld/under_biome_surface");
-    private static final ResourceKey<MaterialRule> DEFAULT_UNDER_BIOME_SURFACE = createKey("overworld/under_biome_surface/default");
-    private static final ResourceKey<MaterialRule> FROZEN_PEAKS_UNDER_SURFACE = createKey("overworld/under_biome_surface/frozen_peaks");
-    private static final ResourceKey<MaterialRule> GROVE_UNDER_SURFACE = createKey("overworld/under_biome_surface/grove");
-    private static final ResourceKey<MaterialRule> JAGGED_PEAKS_UNDER_SURFACE = createKey("overworld/under_biome_surface/jagged_peaks");
-    private static final ResourceKey<MaterialRule> SNOWY_SLOPES_UNDER_SURFACE = createKey("overworld/under_biome_surface/snowy_slopes");
-    private static final ResourceKey<MaterialRule> WINDSWEPT_GRAVELLY_HILLS_UNDER_SURFACE = createKey("overworld/under_biome_surface/windswept_gravelly_hills");
-    private static final ResourceKey<MaterialRule> WINDSWEPT_SAVANNA_UNDER_SURFACE = createKey("overworld/under_biome_surface/windswept_savanna");
-    private static final ResourceKey<MaterialRule> SAND_OR_SANDSTONE_IF_CEILING = createKey("overworld/sand_or_sandstone_if_ceiling");
-    private static final ResourceKey<MaterialRule> GRAVEL_OR_STONE_IF_CEILING = createKey("overworld/gravel_or_stone_if_ceiling");
-    private static final ResourceKey<MaterialRule> POWDER_SNOW_SURFACE = createKey("overworld/powder_snow_surface");
-    private static final ResourceKey<MaterialRule> POWDER_SNOW_UNDER_SURFACE = createKey("overworld/powder_snow_under_surface");
-    private static final MaterialRule AIR = makeStateRule(Blocks.AIR);
-    private static final MaterialRule WHITE_TERRACOTTA = makeStateRule(Blocks.DYED_TERRACOTTA.white());
-    private static final MaterialRule ORANGE_TERRACOTTA = makeStateRule(Blocks.DYED_TERRACOTTA.orange());
-    private static final MaterialRule TERRACOTTA = makeStateRule(Blocks.TERRACOTTA);
-    private static final MaterialRule RED_SAND = makeStateRule(Blocks.RED_SAND);
-    private static final MaterialRule RED_SANDSTONE = makeStateRule(Blocks.RED_SANDSTONE);
-    private static final MaterialRule STONE = makeStateRule(Blocks.STONE);
-    private static final MaterialRule DEEPSLATE = makeStateRule(Blocks.DEEPSLATE);
-    private static final MaterialRule DIRT = makeStateRule(Blocks.DIRT);
-    private static final MaterialRule PODZOL = makeStateRule(Blocks.PODZOL);
-    private static final MaterialRule COARSE_DIRT = makeStateRule(Blocks.COARSE_DIRT);
-    private static final MaterialRule MYCELIUM = makeStateRule(Blocks.MYCELIUM);
-    private static final MaterialRule GRASS_BLOCK = makeStateRule(Blocks.GRASS_BLOCK);
-    private static final MaterialRule CALCITE = makeStateRule(Blocks.CALCITE);
-    private static final MaterialRule GRAVEL = makeStateRule(Blocks.GRAVEL);
-    private static final MaterialRule SAND = makeStateRule(Blocks.SAND);
-    private static final MaterialRule SANDSTONE = makeStateRule(Blocks.SANDSTONE);
-    private static final MaterialRule PACKED_ICE = makeStateRule(Blocks.PACKED_ICE);
-    private static final MaterialRule SNOW_BLOCK = makeStateRule(Blocks.SNOW_BLOCK);
-    private static final MaterialRule MUD = makeStateRule(Blocks.MUD);
-    private static final MaterialRule POWDER_SNOW = makeStateRule(Blocks.POWDER_SNOW);
-    private static final MaterialRule ICE = makeStateRule(Blocks.ICE);
-    private static final MaterialRule WATER = makeStateRule(Blocks.WATER);
-    private static final MaterialRule CINNABAR = makeStateRule(Blocks.CINNABAR);
-    private static final MaterialRule SULFUR = makeStateRule(Blocks.SULFUR);
+    public static final ResourceKey<SurfaceRules.RuleSource> OVERWORLD = createKey("overworld");
+    public static final ResourceKey<SurfaceRules.RuleSource> OVERWORLD_CAVES = createKey("overworld_caves");
+    public static final ResourceKey<SurfaceRules.RuleSource> OVERWORLD_FLOATING_ISLANDS = createKey("overworld_floating_islands");
+    private static final ResourceKey<SurfaceRules.RuleSource> SURFACE = createKey("overworld/surface");
+    private static final ResourceKey<SurfaceRules.RuleSource> UNDERGROUND = createKey("overworld/underground");
+    private static final ResourceKey<SurfaceRules.RuleSource> SULFUR_CAVE_BANDS = createKey("overworld/sulfur_cave_bands");
+    private static final ResourceKey<SurfaceRules.RuleSource> BIOME_SURFACE = createKey("overworld/biome_surface");
+    private static final ResourceKey<SurfaceRules.RuleSource> DAPPLED_FOREST_SURFACE = createKey("overworld/biome_surface/dappled_forest");
+    private static final ResourceKey<SurfaceRules.RuleSource> DEFAULT_BIOME_SURFACE = createKey("overworld/biome_surface/default");
+    private static final ResourceKey<SurfaceRules.RuleSource> DRIPSTONE_CAVES_SURFACE = createKey("overworld/biome_surface/dripstone_caves");
+    private static final ResourceKey<SurfaceRules.RuleSource> FROZEN_PEAKS_SURFACE = createKey("overworld/biome_surface/frozen_peaks");
+    private static final ResourceKey<SurfaceRules.RuleSource> GROVE_SURFACE = createKey("overworld/biome_surface/grove");
+    private static final ResourceKey<SurfaceRules.RuleSource> ICE_SPIKES_SURFACE = createKey("overworld/biome_surface/ice_spikes");
+    private static final ResourceKey<SurfaceRules.RuleSource> JAGGED_PEAKS_SURFACE = createKey("overworld/biome_surface/jagged_peaks");
+    private static final ResourceKey<SurfaceRules.RuleSource> MANGROVE_SWAMP_SURFACE = createKey("overworld/biome_surface/mangrove_swamp");
+    private static final ResourceKey<SurfaceRules.RuleSource> MUSHROOM_FIELDS_SURFACE = createKey("overworld/biome_surface/mushroom_fields");
+    private static final ResourceKey<SurfaceRules.RuleSource> OLD_GROWTH_PINE_TAIGA_SURFACE = createKey("overworld/biome_surface/old_growth_pine_taiga");
+    private static final ResourceKey<SurfaceRules.RuleSource> SNOWY_SLOPES_SURFACE = createKey("overworld/biome_surface/snowy_slopes");
+    private static final ResourceKey<SurfaceRules.RuleSource> STONY_PEAKS_SURFACE = createKey("overworld/biome_surface/stony_peaks");
+    private static final ResourceKey<SurfaceRules.RuleSource> STONY_SHORE_SURFACE = createKey("overworld/biome_surface/stony_shore");
+    private static final ResourceKey<SurfaceRules.RuleSource> SULFUR_CAVES_SURFACE = createKey("overworld/biome_surface/sulfur_caves");
+    private static final ResourceKey<SurfaceRules.RuleSource> WINDSWEPT_GRAVELLY_HILLS_SURFACE = createKey("overworld/biome_surface/windswept_gravelly_hills");
+    private static final ResourceKey<SurfaceRules.RuleSource> WINDSWEPT_HILLS_SURFACE = createKey("overworld/biome_surface/windswept_hills");
+    private static final ResourceKey<SurfaceRules.RuleSource> WINDSWEPT_SAVANNA_SURFACE = createKey("overworld/biome_surface/windswept_savanna");
+    private static final ResourceKey<SurfaceRules.RuleSource> UNDER_BIOME_SURFACE = createKey("overworld/under_biome_surface");
+    private static final ResourceKey<SurfaceRules.RuleSource> DEFAULT_UNDER_BIOME_SURFACE = createKey("overworld/under_biome_surface/default");
+    private static final ResourceKey<SurfaceRules.RuleSource> FROZEN_PEAKS_UNDER_SURFACE = createKey("overworld/under_biome_surface/frozen_peaks");
+    private static final ResourceKey<SurfaceRules.RuleSource> GROVE_UNDER_SURFACE = createKey("overworld/under_biome_surface/grove");
+    private static final ResourceKey<SurfaceRules.RuleSource> JAGGED_PEAKS_UNDER_SURFACE = createKey("overworld/under_biome_surface/jagged_peaks");
+    private static final ResourceKey<SurfaceRules.RuleSource> SNOWY_SLOPES_UNDER_SURFACE = createKey("overworld/under_biome_surface/snowy_slopes");
+    private static final ResourceKey<SurfaceRules.RuleSource> WINDSWEPT_GRAVELLY_HILLS_UNDER_SURFACE = createKey(
+        "overworld/under_biome_surface/windswept_gravelly_hills"
+    );
+    private static final ResourceKey<SurfaceRules.RuleSource> WINDSWEPT_SAVANNA_UNDER_SURFACE = createKey("overworld/under_biome_surface/windswept_savanna");
+    private static final ResourceKey<SurfaceRules.RuleSource> SAND_OR_SANDSTONE_IF_CEILING = createKey("overworld/sand_or_sandstone_if_ceiling");
+    private static final ResourceKey<SurfaceRules.RuleSource> GRAVEL_OR_STONE_IF_CEILING = createKey("overworld/gravel_or_stone_if_ceiling");
+    private static final ResourceKey<SurfaceRules.RuleSource> POWDER_SNOW_SURFACE = createKey("overworld/powder_snow_surface");
+    private static final ResourceKey<SurfaceRules.RuleSource> POWDER_SNOW_UNDER_SURFACE = createKey("overworld/powder_snow_under_surface");
+    private static final SurfaceRules.RuleSource AIR = makeStateRule(Blocks.AIR);
+    private static final SurfaceRules.RuleSource WHITE_TERRACOTTA = makeStateRule(Blocks.DYED_TERRACOTTA.white());
+    private static final SurfaceRules.RuleSource ORANGE_TERRACOTTA = makeStateRule(Blocks.DYED_TERRACOTTA.orange());
+    private static final SurfaceRules.RuleSource TERRACOTTA = makeStateRule(Blocks.TERRACOTTA);
+    private static final SurfaceRules.RuleSource RED_SAND = makeStateRule(Blocks.RED_SAND);
+    private static final SurfaceRules.RuleSource RED_SANDSTONE = makeStateRule(Blocks.RED_SANDSTONE);
+    private static final SurfaceRules.RuleSource STONE = makeStateRule(Blocks.STONE);
+    private static final SurfaceRules.RuleSource DEEPSLATE = makeStateRule(Blocks.DEEPSLATE);
+    private static final SurfaceRules.RuleSource DIRT = makeStateRule(Blocks.DIRT);
+    private static final SurfaceRules.RuleSource PODZOL = makeStateRule(Blocks.PODZOL);
+    private static final SurfaceRules.RuleSource COARSE_DIRT = makeStateRule(Blocks.COARSE_DIRT);
+    private static final SurfaceRules.RuleSource MYCELIUM = makeStateRule(Blocks.MYCELIUM);
+    private static final SurfaceRules.RuleSource GRASS_BLOCK = makeStateRule(Blocks.GRASS_BLOCK);
+    private static final SurfaceRules.RuleSource CALCITE = makeStateRule(Blocks.CALCITE);
+    private static final SurfaceRules.RuleSource GRAVEL = makeStateRule(Blocks.GRAVEL);
+    private static final SurfaceRules.RuleSource SAND = makeStateRule(Blocks.SAND);
+    private static final SurfaceRules.RuleSource SANDSTONE = makeStateRule(Blocks.SANDSTONE);
+    private static final SurfaceRules.RuleSource PACKED_ICE = makeStateRule(Blocks.PACKED_ICE);
+    private static final SurfaceRules.RuleSource SNOW_BLOCK = makeStateRule(Blocks.SNOW_BLOCK);
+    private static final SurfaceRules.RuleSource MUD = makeStateRule(Blocks.MUD);
+    private static final SurfaceRules.RuleSource POWDER_SNOW = makeStateRule(Blocks.POWDER_SNOW);
+    private static final SurfaceRules.RuleSource ICE = makeStateRule(Blocks.ICE);
+    private static final SurfaceRules.RuleSource WATER = makeStateRule(Blocks.WATER);
+    private static final SurfaceRules.RuleSource CINNABAR = makeStateRule(Blocks.CINNABAR);
+    private static final SurfaceRules.RuleSource SULFUR = makeStateRule(Blocks.SULFUR);
 
-    private static ResourceKey<MaterialRule> createKey(String name) {
+    private static ResourceKey<SurfaceRules.RuleSource> createKey(String name) {
         return ResourceKey.create(Registries.MATERIAL_RULE, Identifier.withDefaultNamespace(name));
     }
 
-    private static MaterialRule makeStateRule(Block block) {
-        return MaterialRules.state(block.defaultBlockState());
+    private static SurfaceRules.RuleSource makeStateRule(Block block) {
+        return SurfaceRules.state(block.defaultBlockState());
     }
 
-    public static void bootstrap(BootstrapContext<MaterialRule> context) {
-        HolderGetter<MaterialRule> rules = context.lookup(Registries.MATERIAL_RULE);
+    public static void bootstrap(BootstrapContext<SurfaceRules.RuleSource> context) {
+        HolderGetter<SurfaceRules.RuleSource> rules = context.lookup(Registries.MATERIAL_RULE);
         HolderGetter<Biome> biomes = context.lookup(Registries.BIOME);
-        MaterialRule sulfurCaveBands = MaterialRules.registerAndWrap(
+        SurfaceRules.RuleSource sulfurCaveBands = SurfaceRules.registerAndWrap(
             context,
             SULFUR_CAVE_BANDS,
-            MaterialRules.sequence(
-                MaterialRules.ifTrue(MaterialRules.noiseCondition3d(Noises.SULFUR_CAVE_GRADIENT, -0.4F, -0.1F), CINNABAR),
-                MaterialRules.ifTrue(MaterialRules.noiseCondition3d(Noises.SULFUR_CAVE_GRADIENT, 0.0, 0.4F), SULFUR),
-                MaterialRules.ifTrue(MaterialRules.noiseCondition3d(Noises.SULFUR_CAVE_GRADIENT, 0.4F), CINNABAR)
+            SurfaceRules.sequence(
+                SurfaceRules.ifTrue(SurfaceRules.noiseCondition3d(Noises.SULFUR_CAVE_GRADIENT, -0.4F, -0.1F), CINNABAR),
+                SurfaceRules.ifTrue(SurfaceRules.noiseCondition3d(Noises.SULFUR_CAVE_GRADIENT, 0.0, 0.4F), SULFUR),
+                SurfaceRules.ifTrue(SurfaceRules.noiseCondition3d(Noises.SULFUR_CAVE_GRADIENT, 0.4F), CINNABAR)
             )
         );
-        MaterialRule underground = MaterialRules.registerAndWrap(
+        SurfaceRules.RuleSource underground = SurfaceRules.registerAndWrap(
             context,
             UNDERGROUND,
-            MaterialRules.sequence(
-                MaterialRules.ifTrue(MaterialRules.isBiome(biomes, Biomes.SULFUR_CAVES), sulfurCaveBands),
-                MaterialRules.ifTrue(MaterialRules.verticalGradient("deepslate", VerticalAnchor.absolute(0), VerticalAnchor.absolute(8)), DEEPSLATE)
+            SurfaceRules.sequence(
+                SurfaceRules.ifTrue(SurfaceRules.isBiome(biomes, Biomes.SULFUR_CAVES), sulfurCaveBands),
+                SurfaceRules.ifTrue(SurfaceRules.verticalGradient("deepslate", VerticalAnchor.absolute(0), VerticalAnchor.absolute(8)), DEEPSLATE)
             )
         );
-        List<MaterialRule> oreVeins = registerOreVeins(context);
-        MaterialRule surface = registerSurface(context, sulfurCaveBands);
-        context.register(OVERWORLD, createOverworldLike(rules, true, false, true, surface, underground, oreVeins));
-        context.register(OVERWORLD_CAVES, createOverworldLike(rules, false, true, true, surface, underground, oreVeins));
-        context.register(OVERWORLD_FLOATING_ISLANDS, createOverworldLike(rules, false, false, false, surface, underground, oreVeins));
+        SurfaceRules.RuleSource surface = registerSurface(context, sulfurCaveBands);
+        context.register(OVERWORLD, createOverworldLike(rules, true, false, true, surface, underground));
+        context.register(OVERWORLD_CAVES, createOverworldLike(rules, false, true, true, surface, underground));
+        context.register(OVERWORLD_FLOATING_ISLANDS, createOverworldLike(rules, false, false, false, surface, underground));
     }
 
-    private static MaterialRule registerSurface(BootstrapContext<MaterialRule> context, MaterialRule sulfurCaveBands) {
-        HolderGetter<MaterialCondition> conditions = context.lookup(Registries.MATERIAL_CONDITION);
+    private static SurfaceRules.RuleSource registerSurface(BootstrapContext<SurfaceRules.RuleSource> context, SurfaceRules.RuleSource sulfurCaveBands) {
+        HolderGetter<SurfaceRules.ConditionSource> conditions = context.lookup(Registries.MATERIAL_CONDITION);
         HolderGetter<Biome> biomes = context.lookup(Registries.BIOME);
-        MaterialCondition woodedBadlandsTop = MaterialRules.yBlockCheck(VerticalAnchor.absolute(97), 2);
-        MaterialCondition badlandsTop = MaterialRules.yBlockCheck(VerticalAnchor.absolute(256), 0);
-        MaterialCondition badlandsHeightCondition = MaterialRules.yStartCheck(VerticalAnchor.absolute(63), -1);
-        MaterialCondition badlandsMid = MaterialRules.yStartCheck(VerticalAnchor.absolute(74), 1);
-        MaterialCondition mangroveSwampPuddleLevel = MaterialRules.yBlockCheck(VerticalAnchor.absolute(60), 0);
-        MaterialCondition swampPuddleLevel = MaterialRules.yBlockCheck(VerticalAnchor.absolute(62), 0);
-        MaterialCondition aboveOverworldSeaLevel = MaterialRules.yBlockCheck(VerticalAnchor.absolute(63), 0);
-        MaterialCondition onFloor = MaterialRules.getCondition(conditions, VanillaMaterialConditions.ON_FLOOR);
-        MaterialCondition onCeiling = MaterialRules.getCondition(conditions, VanillaMaterialConditions.ON_CEILING);
-        MaterialCondition underFloor = MaterialRules.getCondition(conditions, VanillaMaterialConditions.UNDER_FLOOR);
-        MaterialCondition notUnderwater = MaterialRules.getCondition(conditions, VanillaMaterialConditions.NOT_UNDERWATER);
-        MaterialCondition notUnderDeepWater = MaterialRules.getCondition(conditions, VanillaMaterialConditions.NOT_UNDER_DEEP_WATER);
-        MaterialCondition hole = MaterialRules.hole();
-        MaterialCondition frozenOcean = MaterialRules.isBiome(biomes, Biomes.FROZEN_OCEAN, Biomes.DEEP_FROZEN_OCEAN);
-        MaterialCondition steep = MaterialRules.steep();
-        MaterialRule defaultBiomeSurface = MaterialRules.registerAndWrap(
-            context, DEFAULT_BIOME_SURFACE, MaterialRules.sequence(MaterialRules.ifTrue(notUnderwater, GRASS_BLOCK), DIRT)
-        );
-        MaterialRule sandOrSandstoneIfCeiling = MaterialRules.registerAndWrap(
-            context, SAND_OR_SANDSTONE_IF_CEILING, MaterialRules.sequence(MaterialRules.ifTrue(onCeiling, SANDSTONE), SAND)
-        );
-        MaterialRule gravelOrStoneIfCeiling = MaterialRules.registerAndWrap(
-            context, GRAVEL_OR_STONE_IF_CEILING, MaterialRules.sequence(MaterialRules.ifTrue(onCeiling, STONE), GRAVEL)
-        );
-        MaterialCondition biomesWithSandAndSandstone = MaterialRules.isBiome(biomes, Biomes.WARM_OCEAN, Biomes.BEACH, Biomes.SNOWY_BEACH);
-        MaterialCondition biomesWithSandAndVeryDeepSandstone = MaterialRules.isBiome(biomes, Biomes.DESERT);
-        MaterialRule commonSurfaceAndUnderRules = MaterialRules.sequence(
-            MaterialRules.ifTrue(
-                MaterialRules.isBiome(biomes, Biomes.STONY_PEAKS),
-                MaterialRules.registerAndWrap(
+        SurfaceRules.ConditionSource woodedBadlandsTop = SurfaceRules.yBlockCheck(VerticalAnchor.absolute(97), 2);
+        SurfaceRules.ConditionSource badlandsTop = SurfaceRules.yBlockCheck(VerticalAnchor.absolute(256), 0);
+        SurfaceRules.ConditionSource badlandsHeightCondition = SurfaceRules.yStartCheck(VerticalAnchor.absolute(63), -1);
+        SurfaceRules.ConditionSource badlandsMid = SurfaceRules.yStartCheck(VerticalAnchor.absolute(74), 1);
+        SurfaceRules.ConditionSource mangroveSwampPuddleLevel = SurfaceRules.yBlockCheck(VerticalAnchor.absolute(60), 0);
+        SurfaceRules.ConditionSource swampPuddleLevel = SurfaceRules.yBlockCheck(VerticalAnchor.absolute(62), 0);
+        SurfaceRules.ConditionSource aboveOverworldSeaLevel = SurfaceRules.yBlockCheck(VerticalAnchor.absolute(63), 0);
+        SurfaceRules.ConditionSource onFloor = SurfaceRules.getCondition(conditions, VanillaMaterialConditions.ON_FLOOR);
+        SurfaceRules.ConditionSource onCeiling = SurfaceRules.getCondition(conditions, VanillaMaterialConditions.ON_CEILING);
+        SurfaceRules.ConditionSource underFloor = SurfaceRules.getCondition(conditions, VanillaMaterialConditions.UNDER_FLOOR);
+        SurfaceRules.ConditionSource notUnderwater = SurfaceRules.getCondition(conditions, VanillaMaterialConditions.NOT_UNDERWATER);
+        SurfaceRules.ConditionSource notUnderDeepWater = SurfaceRules.getCondition(conditions, VanillaMaterialConditions.NOT_UNDER_DEEP_WATER);
+        SurfaceRules.ConditionSource hole = SurfaceRules.hole();
+        SurfaceRules.ConditionSource frozenOcean = SurfaceRules.isBiome(biomes, Biomes.FROZEN_OCEAN, Biomes.DEEP_FROZEN_OCEAN);
+        SurfaceRules.ConditionSource steep = SurfaceRules.steep();
+        SurfaceRules.RuleSource defaultBiomeSurface = SurfaceRules.registerAndWrap(
+            context, DEFAULT_BIOME_SURFACE, SurfaceRules.sequence(SurfaceRules.ifTrue(notUnderwater, GRASS_BLOCK), DIRT)
+        );
+        SurfaceRules.RuleSource sandOrSandstoneIfCeiling = SurfaceRules.registerAndWrap(
+            context, SAND_OR_SANDSTONE_IF_CEILING, SurfaceRules.sequence(SurfaceRules.ifTrue(onCeiling, SANDSTONE), SAND)
+        );
+        SurfaceRules.RuleSource gravelOrStoneIfCeiling = SurfaceRules.registerAndWrap(
+            context, GRAVEL_OR_STONE_IF_CEILING, SurfaceRules.sequence(SurfaceRules.ifTrue(onCeiling, STONE), GRAVEL)
+        );
+        SurfaceRules.ConditionSource biomesWithSandAndSandstone = SurfaceRules.isBiome(biomes, Biomes.WARM_OCEAN, Biomes.BEACH, Biomes.SNOWY_BEACH);
+        SurfaceRules.ConditionSource biomesWithSandAndVeryDeepSandstone = SurfaceRules.isBiome(biomes, Biomes.DESERT);
+        SurfaceRules.RuleSource commonSurfaceAndUnderRules = SurfaceRules.sequence(
+            SurfaceRules.ifTrue(
+                SurfaceRules.isBiome(biomes, Biomes.STONY_PEAKS),
+                SurfaceRules.registerAndWrap(
                     context,
                     STONY_PEAKS_SURFACE,
-                    MaterialRules.sequence(MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.CALCITE, -0.0125, 0.0125), CALCITE), STONE)
+                    SurfaceRules.sequence(SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.CALCITE, -0.0125, 0.0125), CALCITE), STONE)
                 )
             ),
-            MaterialRules.ifTrue(
-                MaterialRules.isBiome(biomes, Biomes.STONY_SHORE),
-                MaterialRules.registerAndWrap(
+            SurfaceRules.ifTrue(
+                SurfaceRules.isBiome(biomes, Biomes.STONY_SHORE),
+                SurfaceRules.registerAndWrap(
                     context,
                     STONY_SHORE_SURFACE,
-                    MaterialRules.sequence(MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.GRAVEL, -0.05, 0.05), gravelOrStoneIfCeiling), STONE)
+                    SurfaceRules.sequence(SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.GRAVEL, -0.05, 0.05), gravelOrStoneIfCeiling), STONE)
                 )
             ),
-            MaterialRules.ifTrue(
-                MaterialRules.isBiome(biomes, Biomes.WINDSWEPT_HILLS),
-                MaterialRules.registerAndWrap(context, WINDSWEPT_HILLS_SURFACE, MaterialRules.ifTrue(surfaceNoiseAbove(1.0), STONE))
-            ),
-            MaterialRules.ifTrue(biomesWithSandAndSandstone, sandOrSandstoneIfCeiling),
-            MaterialRules.ifTrue(biomesWithSandAndVeryDeepSandstone, sandOrSandstoneIfCeiling),
-            MaterialRules.ifTrue(MaterialRules.isBiome(biomes, Biomes.DRIPSTONE_CAVES), MaterialRules.registerAndWrap(context, DRIPSTONE_CAVES_SURFACE, STONE)),
-            MaterialRules.ifTrue(
-                MaterialRules.isBiome(biomes, Biomes.SULFUR_CAVES),
-                MaterialRules.registerAndWrap(context, SULFUR_CAVES_SURFACE, MaterialRules.sequence(sulfurCaveBands, STONE))
-            ),
-            MaterialRules.ifTrue(MaterialRules.isBiome(biomes, Biomes.MANGROVE_SWAMP), MaterialRules.registerAndWrap(context, MANGROVE_SWAMP_SURFACE, MUD))
+            SurfaceRules.ifTrue(
+                SurfaceRules.isBiome(biomes, Biomes.WINDSWEPT_HILLS),
+                SurfaceRules.registerAndWrap(context, WINDSWEPT_HILLS_SURFACE, SurfaceRules.ifTrue(surfaceNoiseAbove(1.0), STONE))
+            ),
+            SurfaceRules.ifTrue(biomesWithSandAndSandstone, sandOrSandstoneIfCeiling),
+            SurfaceRules.ifTrue(biomesWithSandAndVeryDeepSandstone, sandOrSandstoneIfCeiling),
+            SurfaceRules.ifTrue(SurfaceRules.isBiome(biomes, Biomes.DRIPSTONE_CAVES), SurfaceRules.registerAndWrap(context, DRIPSTONE_CAVES_SURFACE, STONE)),
+            SurfaceRules.ifTrue(
+                SurfaceRules.isBiome(biomes, Biomes.SULFUR_CAVES),
+                SurfaceRules.registerAndWrap(context, SULFUR_CAVES_SURFACE, SurfaceRules.sequence(sulfurCaveBands, STONE))
+            ),
+            SurfaceRules.ifTrue(SurfaceRules.isBiome(biomes, Biomes.MANGROVE_SWAMP), SurfaceRules.registerAndWrap(context, MANGROVE_SWAMP_SURFACE, MUD))
         );
-        MaterialRule powderSnowUnderRule = MaterialRules.registerAndWrap(
+        SurfaceRules.RuleSource powderSnowUnderRule = SurfaceRules.registerAndWrap(
             context,
             POWDER_SNOW_UNDER_SURFACE,
-            MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.POWDER_SNOW, 0.45, 0.58), MaterialRules.ifTrue(notUnderwater, POWDER_SNOW))
+            SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.POWDER_SNOW, 0.45, 0.58), SurfaceRules.ifTrue(notUnderwater, POWDER_SNOW))
         );
-        MaterialRule powderSnowSurfaceRule = MaterialRules.registerAndWrap(
+        SurfaceRules.RuleSource powderSnowSurfaceRule = SurfaceRules.registerAndWrap(
             context,
             POWDER_SNOW_SURFACE,
-            MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.POWDER_SNOW, 0.35, 0.6), MaterialRules.ifTrue(notUnderwater, POWDER_SNOW))
+            SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.POWDER_SNOW, 0.35, 0.6), SurfaceRules.ifTrue(notUnderwater, POWDER_SNOW))
         );
-        MaterialRule underBiomeSurfaceRule = MaterialRules.registerAndWrap(
+        SurfaceRules.RuleSource underBiomeSurfaceRule = SurfaceRules.registerAndWrap(
             context,
             UNDER_BIOME_SURFACE,
-            MaterialRules.sequence(
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.FROZEN_PEAKS),
-                    MaterialRules.registerAndWrap(
+            SurfaceRules.sequence(
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.FROZEN_PEAKS),
+                    SurfaceRules.registerAndWrap(
                         context,
                         FROZEN_PEAKS_UNDER_SURFACE,
-                        MaterialRules.sequence(
-                            MaterialRules.ifTrue(steep, PACKED_ICE),
-                            MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.PACKED_ICE, -0.5, 0.2), PACKED_ICE),
-                            MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.ICE, -0.0625, 0.025), ICE),
-                            MaterialRules.ifTrue(notUnderwater, SNOW_BLOCK)
+                        SurfaceRules.sequence(
+                            SurfaceRules.ifTrue(steep, PACKED_ICE),
+                            SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.PACKED_ICE, -0.5, 0.2), PACKED_ICE),
+                            SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.ICE, -0.0625, 0.025), ICE),
+                            SurfaceRules.ifTrue(notUnderwater, SNOW_BLOCK)
                         )
                     )
                 ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.SNOWY_SLOPES),
-                    MaterialRules.registerAndWrap(
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.SNOWY_SLOPES),
+                    SurfaceRules.registerAndWrap(
                         context,
                         SNOWY_SLOPES_UNDER_SURFACE,
-                        MaterialRules.sequence(MaterialRules.ifTrue(steep, STONE), powderSnowUnderRule, MaterialRules.ifTrue(notUnderwater, SNOW_BLOCK))
+                        SurfaceRules.sequence(SurfaceRules.ifTrue(steep, STONE), powderSnowUnderRule, SurfaceRules.ifTrue(notUnderwater, SNOW_BLOCK))
                     )
                 ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.JAGGED_PEAKS), MaterialRules.registerAndWrap(context, JAGGED_PEAKS_UNDER_SURFACE, STONE)
-                ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.GROVE),
-                    MaterialRules.registerAndWrap(context, GROVE_UNDER_SURFACE, MaterialRules.sequence(powderSnowUnderRule, DIRT))
+                SurfaceRules.ifTrue(SurfaceRules.isBiome(biomes, Biomes.JAGGED_PEAKS), SurfaceRules.registerAndWrap(context, JAGGED_PEAKS_UNDER_SURFACE, STONE)),
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.GROVE),
+                    SurfaceRules.registerAndWrap(context, GROVE_UNDER_SURFACE, SurfaceRules.sequence(powderSnowUnderRule, DIRT))
                 ),
                 commonSurfaceAndUnderRules,
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.WINDSWEPT_SAVANNA),
-                    MaterialRules.registerAndWrap(context, WINDSWEPT_SAVANNA_UNDER_SURFACE, MaterialRules.ifTrue(surfaceNoiseAbove(1.75), STONE))
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.WINDSWEPT_SAVANNA),
+                    SurfaceRules.registerAndWrap(context, WINDSWEPT_SAVANNA_UNDER_SURFACE, SurfaceRules.ifTrue(surfaceNoiseAbove(1.75), STONE))
                 ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.WINDSWEPT_GRAVELLY_HILLS),
-                    MaterialRules.registerAndWrap(
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.WINDSWEPT_GRAVELLY_HILLS),
+                    SurfaceRules.registerAndWrap(
                         context,
                         WINDSWEPT_GRAVELLY_HILLS_UNDER_SURFACE,
-                        MaterialRules.sequence(
-                            MaterialRules.ifTrue(surfaceNoiseAbove(2.0), gravelOrStoneIfCeiling),
-                            MaterialRules.ifTrue(surfaceNoiseAbove(1.0), STONE),
-                            MaterialRules.ifTrue(surfaceNoiseAbove(-1.0), DIRT),
+                        SurfaceRules.sequence(
+                            SurfaceRules.ifTrue(surfaceNoiseAbove(2.0), gravelOrStoneIfCeiling),
+                            SurfaceRules.ifTrue(surfaceNoiseAbove(1.0), STONE),
+                            SurfaceRules.ifTrue(surfaceNoiseAbove(-1.0), DIRT),
                             gravelOrStoneIfCeiling
                         )
                     )
                 ),
-                MaterialRules.registerAndWrap(context, DEFAULT_UNDER_BIOME_SURFACE, DIRT)
+                SurfaceRules.registerAndWrap(context, DEFAULT_UNDER_BIOME_SURFACE, DIRT)
             )
         );
-        MaterialRule biomeSurfaceRule = MaterialRules.registerAndWrap(
+        SurfaceRules.RuleSource biomeSurfaceRule = SurfaceRules.registerAndWrap(
             context,
             BIOME_SURFACE,
-            MaterialRules.sequence(
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.FROZEN_PEAKS),
-                    MaterialRules.registerAndWrap(
+            SurfaceRules.sequence(
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.FROZEN_PEAKS),
+                    SurfaceRules.registerAndWrap(
                         context,
                         FROZEN_PEAKS_SURFACE,
-                        MaterialRules.sequence(
-                            MaterialRules.ifTrue(steep, PACKED_ICE),
-                            MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.PACKED_ICE, 0.0, 0.2), PACKED_ICE),
-                            MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.ICE, 0.0, 0.025), ICE),
-                            MaterialRules.ifTrue(notUnderwater, SNOW_BLOCK)
+                        SurfaceRules.sequence(
+                            SurfaceRules.ifTrue(steep, PACKED_ICE),
+                            SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.PACKED_ICE, 0.0, 0.2), PACKED_ICE),
+                            SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.ICE, 0.0, 0.025), ICE),
+                            SurfaceRules.ifTrue(notUnderwater, SNOW_BLOCK)
                         )
                     )
                 ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.SNOWY_SLOPES),
-                    MaterialRules.registerAndWrap(
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.SNOWY_SLOPES),
+                    SurfaceRules.registerAndWrap(
                         context,
                         SNOWY_SLOPES_SURFACE,
-                        MaterialRules.sequence(MaterialRules.ifTrue(steep, STONE), powderSnowSurfaceRule, MaterialRules.ifTrue(notUnderwater, SNOW_BLOCK))
-                    )
-                ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.JAGGED_PEAKS),
-                    MaterialRules.registerAndWrap(
-                        context,
-                        JAGGED_PEAKS_SURFACE,
-                        MaterialRules.sequence(MaterialRules.ifTrue(steep, STONE), MaterialRules.ifTrue(notUnderwater, SNOW_BLOCK))
-                    )
-                ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.GROVE),
-                    MaterialRules.registerAndWrap(
-                        context, GROVE_SURFACE, MaterialRules.sequence(powderSnowSurfaceRule, MaterialRules.ifTrue(notUnderwater, SNOW_BLOCK))
+                        SurfaceRules.sequence(SurfaceRules.ifTrue(steep, STONE), powderSnowSurfaceRule, SurfaceRules.ifTrue(notUnderwater, SNOW_BLOCK))
+                    )
+                ),
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.JAGGED_PEAKS),
+                    SurfaceRules.registerAndWrap(
+                        context, JAGGED_PEAKS_SURFACE, SurfaceRules.sequence(SurfaceRules.ifTrue(steep, STONE), SurfaceRules.ifTrue(notUnderwater, SNOW_BLOCK))
+                    )
+                ),
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.GROVE),
+                    SurfaceRules.registerAndWrap(
+                        context, GROVE_SURFACE, SurfaceRules.sequence(powderSnowSurfaceRule, SurfaceRules.ifTrue(notUnderwater, SNOW_BLOCK))
                     )
                 ),
                 commonSurfaceAndUnderRules,
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.WINDSWEPT_SAVANNA),
-                    MaterialRules.registerAndWrap(
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.WINDSWEPT_SAVANNA),
+                    SurfaceRules.registerAndWrap(
                         context,
                         WINDSWEPT_SAVANNA_SURFACE,
-                        MaterialRules.sequence(MaterialRules.ifTrue(surfaceNoiseAbove(1.75), STONE), MaterialRules.ifTrue(surfaceNoiseAbove(-0.5), COARSE_DIRT))
+                        SurfaceRules.sequence(SurfaceRules.ifTrue(surfaceNoiseAbove(1.75), STONE), SurfaceRules.ifTrue(surfaceNoiseAbove(-0.5), COARSE_DIRT))
                     )
                 ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.WINDSWEPT_GRAVELLY_HILLS),
-                    MaterialRules.registerAndWrap(
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.WINDSWEPT_GRAVELLY_HILLS),
+                    SurfaceRules.registerAndWrap(
                         context,
                         WINDSWEPT_GRAVELLY_HILLS_SURFACE,
-                        MaterialRules.sequence(
-                            MaterialRules.ifTrue(surfaceNoiseAbove(2.0), gravelOrStoneIfCeiling),
-                            MaterialRules.ifTrue(surfaceNoiseAbove(1.0), STONE),
-                            MaterialRules.ifTrue(surfaceNoiseAbove(-1.0), defaultBiomeSurface),
+                        SurfaceRules.sequence(
+                            SurfaceRules.ifTrue(surfaceNoiseAbove(2.0), gravelOrStoneIfCeiling),
+                            SurfaceRules.ifTrue(surfaceNoiseAbove(1.0), STONE),
+                            SurfaceRules.ifTrue(surfaceNoiseAbove(-1.0), defaultBiomeSurface),
                             gravelOrStoneIfCeiling
                         )
                     )
                 ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.OLD_GROWTH_PINE_TAIGA, Biomes.OLD_GROWTH_SPRUCE_TAIGA),
-                    MaterialRules.registerAndWrap(
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.OLD_GROWTH_PINE_TAIGA, Biomes.OLD_GROWTH_SPRUCE_TAIGA),
+                    SurfaceRules.registerAndWrap(
                         context,
                         OLD_GROWTH_PINE_TAIGA_SURFACE,
-                        MaterialRules.sequence(
-                            MaterialRules.ifTrue(surfaceNoiseAbove(1.75), COARSE_DIRT), MaterialRules.ifTrue(surfaceNoiseAbove(-0.95), PODZOL)
-                        )
+                        SurfaceRules.sequence(SurfaceRules.ifTrue(surfaceNoiseAbove(1.75), COARSE_DIRT), SurfaceRules.ifTrue(surfaceNoiseAbove(-0.95), PODZOL))
                     )
                 ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.ICE_SPIKES),
-                    MaterialRules.registerAndWrap(context, ICE_SPIKES_SURFACE, MaterialRules.ifTrue(notUnderwater, SNOW_BLOCK))
-                ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.MUSHROOM_FIELDS), MaterialRules.registerAndWrap(context, MUSHROOM_FIELDS_SURFACE, MYCELIUM)
-                ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.DAPPLED_FOREST),
-                    MaterialRules.registerAndWrap(
-                        context, DAPPLED_FOREST_SURFACE, MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.SMALL_PATCH, 1.2F), COARSE_DIRT)
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.ICE_SPIKES),
+                    SurfaceRules.registerAndWrap(context, ICE_SPIKES_SURFACE, SurfaceRules.ifTrue(notUnderwater, SNOW_BLOCK))
+                ),
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.MUSHROOM_FIELDS), SurfaceRules.registerAndWrap(context, MUSHROOM_FIELDS_SURFACE, MYCELIUM)
+                ),
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.DAPPLED_FOREST),
+                    SurfaceRules.registerAndWrap(
+                        context, DAPPLED_FOREST_SURFACE, SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.SMALL_PATCH, 1.2F), COARSE_DIRT)
                     )
                 ),
                 defaultBiomeSurface
             )
         );
-        MaterialCondition clayBand1 = MaterialRules.noiseCondition2d(Noises.SURFACE, -0.909, -0.5454);
-        MaterialCondition clayBand2 = MaterialRules.noiseCondition2d(Noises.SURFACE, -0.1818, 0.1818);
-        MaterialCondition clayBand3 = MaterialRules.noiseCondition2d(Noises.SURFACE, 0.5454, 0.909);
-        return MaterialRules.registerAndWrap(
+        SurfaceRules.ConditionSource clayBand1 = SurfaceRules.noiseCondition2d(Noises.SURFACE, -0.909, -0.5454);
+        SurfaceRules.ConditionSource clayBand2 = SurfaceRules.noiseCondition2d(Noises.SURFACE, -0.1818, 0.1818);
+        SurfaceRules.ConditionSource clayBand3 = SurfaceRules.noiseCondition2d(Noises.SURFACE, 0.5454, 0.909);
+        return SurfaceRules.registerAndWrap(
             context,
             SURFACE,
-            MaterialRules.sequence(
-                MaterialRules.ifTrue(
+            SurfaceRules.sequence(
+                SurfaceRules.ifTrue(
                     onFloor,
-                    MaterialRules.sequence(
-                        MaterialRules.ifTrue(
-                            MaterialRules.isBiome(biomes, Biomes.WOODED_BADLANDS),
-                            MaterialRules.ifTrue(
+                    SurfaceRules.sequence(
+                        SurfaceRules.ifTrue(
+                            SurfaceRules.isBiome(biomes, Biomes.WOODED_BADLANDS),
+                            SurfaceRules.ifTrue(
                                 woodedBadlandsTop,
-                                MaterialRules.sequence(
-                                    MaterialRules.ifTrue(clayBand1, COARSE_DIRT),
-                                    MaterialRules.ifTrue(clayBand2, COARSE_DIRT),
-                                    MaterialRules.ifTrue(clayBand3, COARSE_DIRT),
+                                SurfaceRules.sequence(
+                                    SurfaceRules.ifTrue(clayBand1, COARSE_DIRT),
+                                    SurfaceRules.ifTrue(clayBand2, COARSE_DIRT),
+                                    SurfaceRules.ifTrue(clayBand3, COARSE_DIRT),
                                     defaultBiomeSurface
                                 )
                             )
                         ),
-                        MaterialRules.ifTrue(
-                            MaterialRules.isBiome(biomes, Biomes.SWAMP),
-                            MaterialRules.ifTrue(
+                        SurfaceRules.ifTrue(
+                            SurfaceRules.isBiome(biomes, Biomes.SWAMP),
+                            SurfaceRules.ifTrue(
                                 swampPuddleLevel,
-                                MaterialRules.ifTrue(
-                                    MaterialRules.not(aboveOverworldSeaLevel), MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.SWAMP, 0.0), WATER)
+                                SurfaceRules.ifTrue(
+                                    SurfaceRules.not(aboveOverworldSeaLevel), SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.SWAMP, 0.0), WATER)
                                 )
                             )
                         ),
-                        MaterialRules.ifTrue(
-                            MaterialRules.isBiome(biomes, Biomes.MANGROVE_SWAMP),
-                            MaterialRules.ifTrue(
+                        SurfaceRules.ifTrue(
+                            SurfaceRules.isBiome(biomes, Biomes.MANGROVE_SWAMP),
+                            SurfaceRules.ifTrue(
                                 mangroveSwampPuddleLevel,
-                                MaterialRules.ifTrue(
-                                    MaterialRules.not(aboveOverworldSeaLevel), MaterialRules.ifTrue(MaterialRules.noiseCondition2d(Noises.SWAMP, 0.0), WATER)
+                                SurfaceRules.ifTrue(
+                                    SurfaceRules.not(aboveOverworldSeaLevel), SurfaceRules.ifTrue(SurfaceRules.noiseCondition2d(Noises.SWAMP, 0.0), WATER)
                                 )
                             )
                         )
                     )
                 ),
-                MaterialRules.ifTrue(
-                    MaterialRules.isBiome(biomes, Biomes.BADLANDS, Biomes.ERODED_BADLANDS, Biomes.WOODED_BADLANDS),
-                    MaterialRules.sequence(
-                        MaterialRules.ifTrue(
+                SurfaceRules.ifTrue(
+                    SurfaceRules.isBiome(biomes, Biomes.BADLANDS, Biomes.ERODED_BADLANDS, Biomes.WOODED_BADLANDS),
+                    SurfaceRules.sequence(
+                        SurfaceRules.ifTrue(
                             onFloor,
-                            MaterialRules.sequence(
-                                MaterialRules.ifTrue(badlandsTop, ORANGE_TERRACOTTA),
-                                MaterialRules.ifTrue(
+                            SurfaceRules.sequence(
+                                SurfaceRules.ifTrue(badlandsTop, ORANGE_TERRACOTTA),
+                                SurfaceRules.ifTrue(
                                     badlandsMid,
-                                    MaterialRules.sequence(
-                                        MaterialRules.ifTrue(clayBand1, TERRACOTTA),
-                                        MaterialRules.ifTrue(clayBand2, TERRACOTTA),
-                                        MaterialRules.ifTrue(clayBand3, TERRACOTTA),
-                                        MaterialRules.bandlands()
+                                    SurfaceRules.sequence(
+                                        SurfaceRules.ifTrue(clayBand1, TERRACOTTA),
+                                        SurfaceRules.ifTrue(clayBand2, TERRACOTTA),
+                                        SurfaceRules.ifTrue(clayBand3, TERRACOTTA),
+                                        SurfaceRules.bandlands()
                                     )
                                 ),
-                                MaterialRules.ifTrue(notUnderwater, MaterialRules.sequence(MaterialRules.ifTrue(onCeiling, RED_SANDSTONE), RED_SAND)),
-                                MaterialRules.ifTrue(MaterialRules.not(hole), ORANGE_TERRACOTTA),
-                                MaterialRules.ifTrue(notUnderDeepWater, WHITE_TERRACOTTA),
+                                SurfaceRules.ifTrue(notUnderwater, SurfaceRules.sequence(SurfaceRules.ifTrue(onCeiling, RED_SANDSTONE), RED_SAND)),
+                                SurfaceRules.ifTrue(SurfaceRules.not(hole), ORANGE_TERRACOTTA),
+                                SurfaceRules.ifTrue(notUnderDeepWater, WHITE_TERRACOTTA),
                                 gravelOrStoneIfCeiling
                             )
                         ),
-                        MaterialRules.ifTrue(
+                        SurfaceRules.ifTrue(
                             badlandsHeightCondition,
-                            MaterialRules.sequence(
-                                MaterialRules.ifTrue(aboveOverworldSeaLevel, MaterialRules.ifTrue(MaterialRules.not(badlandsMid), ORANGE_TERRACOTTA)),
-                                MaterialRules.bandlands()
+                            SurfaceRules.sequence(
+                                SurfaceRules.ifTrue(aboveOverworldSeaLevel, SurfaceRules.ifTrue(SurfaceRules.not(badlandsMid), ORANGE_TERRACOTTA)),
+                                SurfaceRules.bandlands()
                             )
                         ),
-                        MaterialRules.ifTrue(underFloor, MaterialRules.ifTrue(notUnderDeepWater, WHITE_TERRACOTTA))
+                        SurfaceRules.ifTrue(underFloor, SurfaceRules.ifTrue(notUnderDeepWater, WHITE_TERRACOTTA))
                     )
                 ),
-                MaterialRules.ifTrue(
+                SurfaceRules.ifTrue(
                     onFloor,
-                    MaterialRules.ifTrue(
-                        notUnderwater, MaterialRules.sequence(MaterialRules.ifTrue(frozenOcean, MaterialRules.ifTrue(hole, AIR)), biomeSurfaceRule)
+                    SurfaceRules.ifTrue(
+                        notUnderwater, SurfaceRules.sequence(SurfaceRules.ifTrue(frozenOcean, SurfaceRules.ifTrue(hole, AIR)), biomeSurfaceRule)
                     )
                 ),
-                MaterialRules.ifTrue(
+                SurfaceRules.ifTrue(
                     notUnderDeepWater,
-                    MaterialRules.sequence(
-                        MaterialRules.ifTrue(onFloor, MaterialRules.ifTrue(frozenOcean, MaterialRules.ifTrue(hole, WATER))),
-                        MaterialRules.ifTrue(underFloor, underBiomeSurfaceRule),
-                        MaterialRules.ifTrue(
+                    SurfaceRules.sequence(
+                        SurfaceRules.ifTrue(onFloor, SurfaceRules.ifTrue(frozenOcean, SurfaceRules.ifTrue(hole, WATER))),
+                        SurfaceRules.ifTrue(underFloor, underBiomeSurfaceRule),
+                        SurfaceRules.ifTrue(
                             biomesWithSandAndSandstone,
-                            MaterialRules.ifTrue(MaterialRules.getCondition(conditions, VanillaMaterialConditions.DEEP_UNDER_FLOOR), SANDSTONE)
+                            SurfaceRules.ifTrue(SurfaceRules.getCondition(conditions, VanillaMaterialConditions.DEEP_UNDER_FLOOR), SANDSTONE)
                         ),
-                        MaterialRules.ifTrue(
+                        SurfaceRules.ifTrue(
                             biomesWithSandAndVeryDeepSandstone,
-                            MaterialRules.ifTrue(MaterialRules.getCondition(conditions, VanillaMaterialConditions.VERY_DEEP_UNDER_FLOOR), SANDSTONE)
+                            SurfaceRules.ifTrue(SurfaceRules.getCondition(conditions, VanillaMaterialConditions.VERY_DEEP_UNDER_FLOOR), SANDSTONE)
                         )
                     )
                 ),
-                MaterialRules.ifTrue(
+                SurfaceRules.ifTrue(
                     onFloor,
-                    MaterialRules.sequence(
-                        MaterialRules.ifTrue(MaterialRules.isBiome(biomes, Biomes.FROZEN_PEAKS, Biomes.JAGGED_PEAKS), STONE),
-                        MaterialRules.ifTrue(
-                            MaterialRules.isBiome(biomes, Biomes.WARM_OCEAN, Biomes.LUKEWARM_OCEAN, Biomes.DEEP_LUKEWARM_OCEAN), sandOrSandstoneIfCeiling
+                    SurfaceRules.sequence(
+                        SurfaceRules.ifTrue(SurfaceRules.isBiome(biomes, Biomes.FROZEN_PEAKS, Biomes.JAGGED_PEAKS), STONE),
+                        SurfaceRules.ifTrue(
+                            SurfaceRules.isBiome(biomes, Biomes.WARM_OCEAN, Biomes.LUKEWARM_OCEAN, Biomes.DEEP_LUKEWARM_OCEAN), sandOrSandstoneIfCeiling
                         ),
                         gravelOrStoneIfCeiling
                     )
@@ -439,50 +_,30 @@
         );
     }
 
-    private static MaterialRule createOverworldLike(
-        HolderGetter<MaterialRule> rules,
+    private static SurfaceRules.RuleSource createOverworldLike(
+        HolderGetter<SurfaceRules.RuleSource> rules,
         boolean doPreliminarySurfaceCheck,
         boolean bedrockRoof,
         boolean bedrockFloor,
-        MaterialRule mainRuleCloseToSurface,
-        MaterialRule underground,
-        List<MaterialRule> oreVeins
+        SurfaceRules.RuleSource mainRuleCloseToSurface,
+        SurfaceRules.RuleSource underground
     ) {
-        Builder<MaterialRule> builder = ImmutableList.builder();
+        Builder<SurfaceRules.RuleSource> builder = ImmutableList.builder();
         if (bedrockRoof) {
-            builder.add(MaterialRules.getRule(rules, VanillaMaterialRules.BEDROCK_ROOF));
+            builder.add(SurfaceRules.getRule(rules, VanillaMaterialRules.BEDROCK_ROOF));
         }
 
         if (bedrockFloor) {
-            builder.add(MaterialRules.getRule(rules, VanillaMaterialRules.BEDROCK_FLOOR));
+            builder.add(SurfaceRules.getRule(rules, VanillaMaterialRules.BEDROCK_FLOOR));
         }
 
-        builder.addAll(oreVeins);
-        MaterialRule ruleAbovePreliminarySurface = MaterialRules.ifTrue(MaterialRules.abovePreliminarySurface(), mainRuleCloseToSurface);
+        SurfaceRules.RuleSource ruleAbovePreliminarySurface = SurfaceRules.ifTrue(SurfaceRules.abovePreliminarySurface(), mainRuleCloseToSurface);
         builder.add(doPreliminarySurfaceCheck ? ruleAbovePreliminarySurface : mainRuleCloseToSurface);
         builder.add(underground);
-        return MaterialRules.sequence(builder.build());
-    }
-
-    private static MaterialCondition surfaceNoiseAbove(double threshold) {
-        return MaterialRules.noiseCondition2d(Noises.SURFACE, threshold / 8.25, Double.MAX_VALUE);
-    }
-
-    private static List<MaterialRule> registerOreVeins(BootstrapContext<MaterialRule> context) {
-        HolderGetter<DensityFunction> functions = context.lookup(Registries.DENSITY_FUNCTION);
-        DensityFunction richness = NoiseRouterData.getFunction(functions, NoiseRouterData.ORE_VEIN_RICHNESS);
-        DensityFunction gap = NoiseRouterData.getFunction(functions, NoiseRouterData.ORE_VEIN_GAP);
-        return List.of(
-            MaterialRules.registerAndWrap(
-                context,
-                COPPER_ORE_VEIN,
-                OreVeinRule.VeinType.COPPER.create(NoiseRouterData.getFunction(functions, NoiseRouterData.ORE_VEIN_COPPER_DENSITY), richness, gap)
-            ),
-            MaterialRules.registerAndWrap(
-                context,
-                IRON_ORE_VEIN,
-                OreVeinRule.VeinType.IRON.create(NoiseRouterData.getFunction(functions, NoiseRouterData.ORE_VEIN_IRON_DENSITY), richness, gap)
-            )
-        );
+        return SurfaceRules.sequence(builder.build().toArray(SurfaceRules.RuleSource[]::new));
+    }
+
+    private static SurfaceRules.ConditionSource surfaceNoiseAbove(double threshold) {
+        return SurfaceRules.noiseCondition2d(Noises.SURFACE, threshold / 8.25, Double.MAX_VALUE);
     }
 }
