--- a/net/minecraft/data/worldgen/material/VanillaMaterialRules.java
+++ b/net/minecraft/data/worldgen/material/VanillaMaterialRules.java
@@ -6,38 +_,36 @@
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.levelgen.SurfaceRules;
 import net.minecraft.world.level.levelgen.VerticalAnchor;
-import net.minecraft.world.level.levelgen.material.MaterialRules;
-import net.minecraft.world.level.levelgen.material.rule.MaterialRule;
 
 public class VanillaMaterialRules {
-    public static final ResourceKey<MaterialRule> BEDROCK_FLOOR = createKey("bedrock_floor");
-    public static final ResourceKey<MaterialRule> BEDROCK_ROOF = createKey("bedrock_roof");
-    private static final MaterialRule BEDROCK = makeStateRule(Blocks.BEDROCK);
+    public static final ResourceKey<SurfaceRules.RuleSource> BEDROCK_FLOOR = createKey("bedrock_floor");
+    public static final ResourceKey<SurfaceRules.RuleSource> BEDROCK_ROOF = createKey("bedrock_roof");
+    private static final SurfaceRules.RuleSource BEDROCK = makeStateRule(Blocks.BEDROCK);
 
-    private static ResourceKey<MaterialRule> createKey(String name) {
+    private static ResourceKey<SurfaceRules.RuleSource> createKey(String name) {
         return ResourceKey.create(Registries.MATERIAL_RULE, Identifier.withDefaultNamespace(name));
     }
 
-    private static MaterialRule makeStateRule(Block block) {
-        return MaterialRules.state(block.defaultBlockState());
+    private static SurfaceRules.RuleSource makeStateRule(Block block) {
+        return SurfaceRules.state(block.defaultBlockState());
     }
 
-    public static void bootstrap(BootstrapContext<MaterialRule> context) {
+    public static void bootstrap(BootstrapContext<SurfaceRules.RuleSource> context) {
         context.register(
-            BEDROCK_FLOOR,
-            MaterialRules.ifTrue(MaterialRules.verticalGradient("bedrock_floor", VerticalAnchor.bottom(), VerticalAnchor.aboveBottom(5)), BEDROCK)
+            BEDROCK_FLOOR, SurfaceRules.ifTrue(SurfaceRules.verticalGradient("bedrock_floor", VerticalAnchor.bottom(), VerticalAnchor.aboveBottom(5)), BEDROCK)
         );
         context.register(
             BEDROCK_ROOF,
-            MaterialRules.ifTrue(MaterialRules.not(MaterialRules.verticalGradient("bedrock_roof", VerticalAnchor.belowTop(5), VerticalAnchor.top())), BEDROCK)
+            SurfaceRules.ifTrue(SurfaceRules.not(SurfaceRules.verticalGradient("bedrock_roof", VerticalAnchor.belowTop(5), VerticalAnchor.top())), BEDROCK)
         );
         OverworldMaterialRules.bootstrap(context);
         NetherMaterialRules.bootstrap(context);
         EndMaterialRules.bootstrap(context);
     }
 
-    public static MaterialRule air() {
-        return MaterialRules.state(Blocks.AIR.defaultBlockState());
+    public static SurfaceRules.RuleSource air() {
+        return SurfaceRules.state(Blocks.AIR.defaultBlockState());
     }
 }
