--- a/net/minecraft/client/renderer/entity/AbstractZombieRenderer.java
+++ b/net/minecraft/client/renderer/entity/AbstractZombieRenderer.java
@@ -1,15 +_,20 @@
 package net.minecraft.client.renderer.entity;
 
 import net.minecraft.client.model.HumanoidModel;
+import net.minecraft.client.model.monster.zombie.ZombieModel;
 import net.minecraft.client.renderer.entity.layers.HumanoidArmorLayer;
 import net.minecraft.client.renderer.entity.state.ZombieRenderState;
+import net.minecraft.core.component.DataComponents;
 import net.minecraft.resources.Identifier;
+import net.minecraft.world.entity.HumanoidArm;
 import net.minecraft.world.entity.monster.zombie.Zombie;
+import net.minecraft.world.item.SwingAnimationType;
+import net.minecraft.world.item.component.SwingAnimation;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public abstract class AbstractZombieRenderer<T extends Zombie, S extends ZombieRenderState, M extends HumanoidModel<S>> extends HumanoidMobRenderer<T, S, M> {
+public abstract class AbstractZombieRenderer<T extends Zombie, S extends ZombieRenderState, M extends ZombieModel<S>> extends HumanoidMobRenderer<T, S, M> {
     private static final Identifier ZOMBIE_LOCATION = Identifier.withDefaultNamespace("textures/entity/zombie/zombie.png");
     private static final Identifier BABY_ZOMBIE_LOCATION = Identifier.withDefaultNamespace("textures/entity/zombie/zombie_baby.png");
 
@@ -30,5 +_,10 @@
 
     protected boolean isShaking(S state) {
         return super.isShaking(state) || state.isConverting;
+    }
+
+    protected HumanoidModel.ArmPose getArmPose(T mob, HumanoidArm arm) {
+        SwingAnimation otherAnim = mob.getItemHeldByArm(arm.getOpposite()).get(DataComponents.SWING_ANIMATION);
+        return otherAnim != null && otherAnim.type() == SwingAnimationType.STAB ? HumanoidModel.ArmPose.SPEAR : super.getArmPose(mob, arm);
     }
 }
