--- a/net/minecraft/client/renderer/entity/AbstractBoatRenderer.java
+++ b/net/minecraft/client/renderer/entity/AbstractBoatRenderer.java
@@ -27,18 +_,18 @@
     public void submit(BoatRenderState state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
         poseStack.pushPose();
         poseStack.translate(0.0F, 0.375F, 0.0F);
-        poseStack.rotateDegrees(Axis.YP, 180.0F - state.yRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(180.0F - state.yRot));
         float hurt = state.hurtTime;
         if (hurt > 0.0F) {
-            poseStack.rotateDegrees(Axis.XP, Mth.sin(hurt) * hurt * state.damageTime / 10.0F * state.hurtDir);
+            poseStack.mulPose(Axis.XP.rotationDegrees(Mth.sin(hurt) * hurt * state.damageTime / 10.0F * state.hurtDir));
         }
 
         if (!state.isUnderWater && !Mth.equal(state.bubbleAngle, 0.0F)) {
-            poseStack.rotate(new Quaternionf().setAngleAxis(state.bubbleAngle * (float) (Math.PI / 180.0), 1.0F, 0.0F, 1.0F));
+            poseStack.mulPose(new Quaternionf().setAngleAxis(state.bubbleAngle * (float) (Math.PI / 180.0), 1.0F, 0.0F, 1.0F));
         }
 
         poseStack.scale(-1.0F, -1.0F, 1.0F);
-        poseStack.rotateDegrees(Axis.YP, 90.0F);
+        poseStack.mulPose(Axis.YP.rotationDegrees(90.0F));
         submitNodeCollector.submitModel(this.model(), state, poseStack, this.texture, state.lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
         this.submitTypeAdditions(state, poseStack, submitNodeCollector, state.lightCoords);
         poseStack.popPose();
