--- a/net/minecraft/client/renderer/entity/AbstractMinecartRenderer.java
+++ b/net/minecraft/client/renderer/entity/AbstractMinecartRenderer.java
@@ -53,7 +_,7 @@
 
         float hurt = state.hurtTime;
         if (hurt > 0.0F) {
-            poseStack.rotateDegrees(Axis.XP, Mth.sin(hurt) * hurt * state.damageTime / 10.0F * state.hurtDir);
+            poseStack.mulPose(Axis.XP.rotationDegrees(Mth.sin(hurt) * hurt * state.damageTime / 10.0F * state.hurtDir));
         }
 
         BlockModelRenderState displayBlockModel = state.displayBlockModel;
@@ -61,7 +_,7 @@
             poseStack.pushPose();
             poseStack.scale(0.75F, 0.75F, 0.75F);
             poseStack.translate(-0.5F, (state.displayOffset - 8) / 16.0F, 0.5F);
-            poseStack.rotateDegrees(Axis.YP, 90.0F);
+            poseStack.mulPose(Axis.YP.rotationDegrees(90.0F));
             this.submitMinecartContents(state, displayBlockModel, poseStack, submitNodeCollector, state.lightCoords);
             poseStack.popPose();
         }
@@ -72,8 +_,8 @@
     }
 
     private static <S extends MinecartRenderState> void newRender(S state, PoseStack poseStack) {
-        poseStack.rotateDegrees(Axis.YP, state.yRot);
-        poseStack.rotateDegrees(Axis.ZP, -state.xRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(state.yRot));
+        poseStack.mulPose(Axis.ZP.rotationDegrees(-state.xRot));
         poseStack.translate(0.0F, 0.375F, 0.0F);
     }
 
@@ -96,8 +_,8 @@
         }
 
         poseStack.translate(0.0F, 0.375F, 0.0F);
-        poseStack.rotateDegrees(Axis.YP, 180.0F - rotation);
-        poseStack.rotateDegrees(Axis.ZP, -xRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(180.0F - rotation));
+        poseStack.mulPose(Axis.ZP.rotationDegrees(-xRot));
     }
 
     public void extractRenderState(T entity, S state, float partialTicks) {
@@ -162,8 +_,8 @@
         blockModel.submit(poseStack, submitNodeCollector, lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
     }
 
-    protected AABB getBoundingBoxForCulling(T entity, float partialTicks) {
-        AABB aabb = super.getBoundingBoxForCulling(entity, partialTicks);
+    protected AABB getBoundingBoxForCulling(T entity) {
+        AABB aabb = super.getBoundingBoxForCulling(entity);
         return !entity.getDisplayBlockState().isAir() ? aabb.expandTowards(0.0, entity.getDisplayOffset() * 0.75F / 16.0F, 0.0) : aabb;
     }
 
