--- a/net/minecraft/client/renderer/entity/ArmorStandRenderer.java
+++ b/net/minecraft/client/renderer/entity/ArmorStandRenderer.java
@@ -74,9 +_,9 @@
     }
 
     protected void setupRotations(ArmorStandRenderState state, PoseStack poseStack, float bodyRot, float entityScale) {
-        poseStack.rotateDegrees(Axis.YP, 180.0F - bodyRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(180.0F - bodyRot));
         if (state.wiggle < 5.0F) {
-            poseStack.rotateDegrees(Axis.YP, Mth.sin(state.wiggle / 1.5F * (float) Math.PI) * 3.0F);
+            poseStack.mulPose(Axis.YP.rotationDegrees(Mth.sin(state.wiggle / 1.5F * (float) Math.PI) * 3.0F));
         }
     }
 
