--- a/net/minecraft/client/renderer/entity/ArrowRenderer.java
+++ b/net/minecraft/client/renderer/entity/ArrowRenderer.java
@@ -24,8 +_,8 @@
 
     public void submit(S state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
         poseStack.pushPose();
-        poseStack.rotateDegrees(Axis.YP, state.yRot - 90.0F);
-        poseStack.rotateDegrees(Axis.ZP, state.xRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(state.yRot - 90.0F));
+        poseStack.mulPose(Axis.ZP.rotationDegrees(state.xRot));
         submitNodeCollector.submitModel(
             this.model, state, poseStack, this.getTextureLocation(state), state.lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor
         );
