--- a/net/minecraft/client/renderer/entity/CatRenderer.java
+++ b/net/minecraft/client/renderer/entity/CatRenderer.java
@@ -47,7 +_,7 @@
         float lieDownAmount = state.lieDownAmount;
         if (lieDownAmount > 0.0F) {
             poseStack.translate(0.4F * lieDownAmount, 0.15F * lieDownAmount, 0.1F * lieDownAmount);
-            poseStack.rotateDegrees(Axis.ZP, Mth.rotLerp(lieDownAmount, 0.0F, 90.0F));
+            poseStack.mulPose(Axis.ZP.rotationDegrees(Mth.rotLerp(lieDownAmount, 0.0F, 90.0F)));
             if (state.isLyingOnTopOfSleepingPlayer) {
                 poseStack.translate(0.15F * lieDownAmount, 0.0F, 0.0F);
             }
