--- a/net/minecraft/client/renderer/entity/CodRenderer.java
+++ b/net/minecraft/client/renderer/entity/CodRenderer.java
@@ -32,10 +_,10 @@
     protected void setupRotations(LivingEntityRenderState state, PoseStack poseStack, float bodyRot, float entityScale) {
         super.setupRotations(state, poseStack, bodyRot, entityScale);
         float bodyZRot = 4.3F * Mth.sin(0.6F * state.ageInTicks);
-        poseStack.rotateDegrees(Axis.YP, bodyZRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(bodyZRot));
         if (!state.isInWater) {
             poseStack.translate(0.1F, 0.1F, -0.1F);
-            poseStack.rotateDegrees(Axis.ZP, 90.0F);
+            poseStack.mulPose(Axis.ZP.rotationDegrees(90.0F));
         }
     }
 }
