--- a/net/minecraft/client/renderer/entity/CushionRenderer.java
+++ b/net/minecraft/client/renderer/entity/CushionRenderer.java
@@ -39,8 +_,8 @@
 
     public void submit(CushionRenderState state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
         poseStack.pushPose();
-        poseStack.rotateDegrees(Axis.YP, 180.0F - state.direction.toYRot());
-        poseStack.rotateDegrees(Axis.XP, 180.0F);
+        poseStack.mulPose(Axis.YP.rotationDegrees(state.direction.toYRot()));
+        poseStack.mulPose(Axis.XP.rotationDegrees(180.0F));
         poseStack.translate(0.0, -0.25, 0.0);
         submitNodeCollector.submitModel(
             this.model, state, poseStack, this.model.renderType(state.texture), state.lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor
