--- a/net/minecraft/client/renderer/entity/DragonFireballRenderer.java
+++ b/net/minecraft/client/renderer/entity/DragonFireballRenderer.java
@@ -31,7 +_,7 @@
     public void submit(EntityRenderState state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
         poseStack.pushPose();
         poseStack.scale(2.0F, 2.0F, 2.0F);
-        poseStack.rotate(camera.orientation);
+        poseStack.mulPose(camera.orientation);
         submitNodeCollector.submitCustomGeometry(poseStack, RENDER_TYPE, (pose, buffer) -> buildQuad(state, pose, buffer, -1));
         if (state.outlineColor != 0 && RENDER_TYPE.outline().isPresent()) {
             submitNodeCollector.submitCustomGeometry(
