--- a/net/minecraft/client/renderer/entity/EndCrystalRenderer.java
+++ b/net/minecraft/client/renderer/entity/EndCrystalRenderer.java
@@ -68,7 +_,7 @@
         }
     }
 
-    public boolean shouldRender(EndCrystal entity, Frustum culler, double camX, double camY, double camZ, float partialTicks) {
-        return super.shouldRender(entity, culler, camX, camY, camZ, partialTicks) || entity.getBeamTarget() != null;
+    public boolean shouldRender(EndCrystal entity, Frustum culler, double camX, double camY, double camZ) {
+        return super.shouldRender(entity, culler, camX, camY, camZ) || entity.getBeamTarget() != null;
     }
 }
