--- a/net/minecraft/client/renderer/entity/EndermanRenderer.java
+++ b/net/minecraft/client/renderer/entity/EndermanRenderer.java
@@ -9,14 +_,14 @@
 import net.minecraft.client.renderer.entity.state.EndermanRenderState;
 import net.minecraft.resources.Identifier;
 import net.minecraft.util.RandomSource;
-import net.minecraft.world.entity.monster.Enderman;
+import net.minecraft.world.entity.monster.EnderMan;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec3;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public class EndermanRenderer extends MobRenderer<Enderman, EndermanRenderState, EndermanModel<EndermanRenderState>> {
+public class EndermanRenderer extends MobRenderer<EnderMan, EndermanRenderState, EndermanModel<EndermanRenderState>> {
     public static final BlockDisplayContext BLOCK_DISPLAY_CONTEXT = BlockDisplayContext.create();
     private static final Identifier ENDERMAN_LOCATION = Identifier.withDefaultNamespace("textures/entity/enderman/enderman.png");
     private final RandomSource random = RandomSource.create();
@@ -47,7 +_,7 @@
         return new EndermanRenderState();
     }
 
-    public void extractRenderState(Enderman entity, EndermanRenderState state, float partialTicks) {
+    public void extractRenderState(EnderMan entity, EndermanRenderState state, float partialTicks) {
         super.extractRenderState(entity, state, partialTicks);
         HumanoidMobRenderer.extractHumanoidRenderState(entity, state, partialTicks, this.itemModelResolver);
         state.isCreepy = entity.isCreepy();
