--- a/net/minecraft/client/renderer/entity/EnderDragonRenderer.java
+++ b/net/minecraft/client/renderer/entity/EnderDragonRenderer.java
@@ -49,8 +_,8 @@
         poseStack.pushPose();
         float yr = state.getHistoricalPos(7).yRot();
         float rot2 = (float)(state.getHistoricalPos(5).y() - state.getHistoricalPos(10).y());
-        poseStack.rotateDegrees(Axis.YP, -yr);
-        poseStack.rotateDegrees(Axis.XP, rot2 * 10.0F);
+        poseStack.mulPose(Axis.YP.rotationDegrees(-yr));
+        poseStack.mulPose(Axis.XP.rotationDegrees(rot2 * 10.0F));
         poseStack.translate(0.0F, 0.0F, 1.0F);
         poseStack.scale(-1.0F, -1.0F, 1.0F);
         poseStack.translate(0.0F, -1.501F, 0.0F);
@@ -141,8 +_,8 @@
         float length = Mth.sqrt(deltaX * deltaX + deltaY * deltaY + deltaZ * deltaZ);
         poseStack.pushPose();
         poseStack.translate(0.0F, 2.0F, 0.0F);
-        poseStack.rotate(Axis.YP, (float)(-Math.atan2(deltaZ, deltaX)) - (float) (Math.PI / 2));
-        poseStack.rotate(Axis.XP, (float)(-Math.atan2(horizontalLength, deltaY)) - (float) (Math.PI / 2));
+        poseStack.mulPose(Axis.YP.rotation((float)(-Math.atan2(deltaZ, deltaX)) - (float) (Math.PI / 2)));
+        poseStack.mulPose(Axis.XP.rotation((float)(-Math.atan2(horizontalLength, deltaY)) - (float) (Math.PI / 2)));
         float v0 = 0.0F - timeInTicks * 0.01F;
         float v1 = length / 32.0F - timeInTicks * 0.01F;
         submitNodeCollector.submitCustomGeometry(
