--- a/net/minecraft/client/renderer/entity/EntityRenderers.java
+++ b/net/minecraft/client/renderer/entity/EntityRenderers.java
@@ -25,7 +_,7 @@
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final Map<EntityType<?>, EntityRendererProvider<?>> PROVIDERS = new Object2ObjectOpenHashMap<>();
 
-    public static <T extends Entity> void register(EntityType<? extends T> type, EntityRendererProvider<T> renderer) {
+    private static <T extends Entity> void register(EntityType<? extends T> type, EntityRendererProvider<T> renderer) {
         PROVIDERS.put(type, renderer);
     }
 
