--- a/net/minecraft/client/renderer/entity/EvokerFangsRenderer.java
+++ b/net/minecraft/client/renderer/entity/EvokerFangsRenderer.java
@@ -27,7 +_,7 @@
         float biteProgress = state.biteProgress;
         if (biteProgress != 0.0F) {
             poseStack.pushPose();
-            poseStack.rotateDegrees(Axis.YP, 90.0F - state.yRot);
+            poseStack.mulPose(Axis.YP.rotationDegrees(90.0F - state.yRot));
             poseStack.scale(-1.0F, -1.0F, 1.0F);
             poseStack.translate(0.0F, -1.501F, 0.0F);
             submitNodeCollector.submitModel(this.model, state, poseStack, TEXTURE_LOCATION, state.lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
