--- a/net/minecraft/client/renderer/entity/ExperienceOrbRenderer.java
+++ b/net/minecraft/client/renderer/entity/ExperienceOrbRenderer.java
@@ -46,7 +_,7 @@
         int gc = 255;
         int bc = (int)((Mth.sin(rr + (float) (Math.PI * 4.0 / 3.0)) + 1.0F) * 0.1F * 255.0F);
         poseStack.translate(0.0F, 0.1F, 0.0F);
-        poseStack.rotate(camera.orientation);
+        poseStack.mulPose(camera.orientation);
         float s = 0.3F;
         poseStack.scale(0.3F, 0.3F, 0.3F);
         submitNodeCollector.submitCustomGeometry(poseStack, RENDER_TYPE, (pose, buffer) -> {
