--- a/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
+++ b/net/minecraft/client/renderer/entity/FallingBlockRenderer.java
@@ -20,10 +_,8 @@
         this.shadowRadius = 0.5F;
     }
 
-    public boolean shouldRender(FallingBlockEntity entity, Frustum culler, double camX, double camY, double camZ, float partialTicks) {
-        return !super.shouldRender(entity, culler, camX, camY, camZ, partialTicks)
-            ? false
-            : entity.getBlockState() != entity.level().getBlockState(entity.blockPosition());
+    public boolean shouldRender(FallingBlockEntity entity, Frustum culler, double camX, double camY, double camZ) {
+        return !super.shouldRender(entity, culler, camX, camY, camZ) ? false : entity.getBlockState() != entity.level().getBlockState(entity.blockPosition());
     }
 
     public void submit(FallingBlockRenderState state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
