--- a/net/minecraft/client/renderer/entity/FireworkEntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/FireworkEntityRenderer.java
@@ -23,11 +_,11 @@
 
     public void submit(FireworkRocketRenderState state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
         poseStack.pushPose();
-        poseStack.rotate(camera.orientation);
+        poseStack.mulPose(camera.orientation);
         if (state.isShotAtAngle) {
-            poseStack.rotateDegrees(Axis.ZP, 180.0F);
-            poseStack.rotateDegrees(Axis.YP, 180.0F);
-            poseStack.rotateDegrees(Axis.XP, 90.0F);
+            poseStack.mulPose(Axis.ZP.rotationDegrees(180.0F));
+            poseStack.mulPose(Axis.YP.rotationDegrees(180.0F));
+            poseStack.mulPose(Axis.XP.rotationDegrees(90.0F));
         }
 
         state.item.submit(poseStack, submitNodeCollector, state.lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
