--- a/net/minecraft/client/renderer/entity/FoxRenderer.java
+++ b/net/minecraft/client/renderer/entity/FoxRenderer.java
@@ -50,7 +_,7 @@
     protected void setupRotations(FoxRenderState state, PoseStack poseStack, float bodyRot, float entityScale) {
         super.setupRotations(state, poseStack, bodyRot, entityScale);
         if (state.isPouncing || state.isFaceplanted) {
-            poseStack.rotateDegrees(Axis.XP, -state.xRot);
+            poseStack.mulPose(Axis.XP.rotationDegrees(-state.xRot));
         }
     }
 
