--- a/net/minecraft/client/renderer/entity/GuardianRenderer.java
+++ b/net/minecraft/client/renderer/entity/GuardianRenderer.java
@@ -39,8 +_,8 @@
         super(context, new GuardianModel(context.bakeLayer(modelId)), shadow);
     }
 
-    public boolean shouldRender(Guardian entity, Frustum culler, double camX, double camY, double camZ, float partialTicks) {
-        if (super.shouldRender(entity, culler, camX, camY, camZ, partialTicks)) {
+    public boolean shouldRender(Guardian entity, Frustum culler, double camX, double camY, double camZ) {
+        if (super.shouldRender(entity, culler, camX, camY, camZ)) {
             return true;
         }
 
@@ -80,8 +_,8 @@
         beamVector = beamVector.normalize();
         float xRot = (float)Math.acos(beamVector.y);
         float yRot = (float) (Math.PI / 2) - (float)Math.atan2(beamVector.z, beamVector.x);
-        poseStack.rotateDegrees(Axis.YP, yRot * (180.0F / (float)Math.PI));
-        poseStack.rotateDegrees(Axis.XP, xRot * (180.0F / (float)Math.PI));
+        poseStack.mulPose(Axis.YP.rotationDegrees(yRot * (180.0F / (float)Math.PI)));
+        poseStack.mulPose(Axis.XP.rotationDegrees(xRot * (180.0F / (float)Math.PI)));
         float rot = timeInTicks * 0.05F * -1.5F;
         float colorScale = scale * scale;
         int red = 64 + (int)(colorScale * 191.0F);
