--- a/net/minecraft/client/renderer/entity/HappyGhastRenderer.java
+++ b/net/minecraft/client/renderer/entity/HappyGhastRenderer.java
@@ -48,8 +_,8 @@
         return new HappyGhastRenderState();
     }
 
-    protected AABB getBoundingBoxForCulling(HappyGhast entity, float partialTicks) {
-        AABB aabb = super.getBoundingBoxForCulling(entity, partialTicks);
+    protected AABB getBoundingBoxForCulling(HappyGhast entity) {
+        AABB aabb = super.getBoundingBoxForCulling(entity);
         float height = entity.getBbHeight();
         return aabb.setMinY(aabb.minY - height / 2.0F);
     }
