--- a/net/minecraft/client/renderer/entity/IllagerRenderer.java
+++ b/net/minecraft/client/renderer/entity/IllagerRenderer.java
@@ -26,6 +_,7 @@
             ? CrossbowItem.getChargeDuration(entity.getUseItem(), entity)
             : 0;
         state.ticksUsingItem = entity.getTicksUsingItem(partialTicks);
+        state.attackAnim = entity.getAttackAnim(partialTicks);
         state.isAggressive = entity.isAggressive();
     }
 }
