--- a/net/minecraft/client/renderer/entity/IllusionerRenderer.java
+++ b/net/minecraft/client/renderer/entity/IllusionerRenderer.java
@@ -74,7 +_,7 @@
         return true;
     }
 
-    protected AABB getBoundingBoxForCulling(Illusioner entity, float partialTicks) {
-        return super.getBoundingBoxForCulling(entity, partialTicks).inflate(3.0, 0.0, 3.0);
+    protected AABB getBoundingBoxForCulling(Illusioner entity) {
+        return super.getBoundingBoxForCulling(entity).inflate(3.0, 0.0, 3.0);
     }
 }
