--- a/net/minecraft/client/renderer/entity/IronGolemRenderer.java
+++ b/net/minecraft/client/renderer/entity/IronGolemRenderer.java
@@ -55,7 +_,7 @@
             float p = 13.0F;
             float wp = state.walkAnimationPos + 6.0F;
             float triangleWave = (Math.abs(wp % 13.0F - 6.5F) - 3.25F) / 3.25F;
-            poseStack.rotateDegrees(Axis.ZP, 6.5F * triangleWave);
+            poseStack.mulPose(Axis.ZP.rotationDegrees(6.5F * triangleWave));
         }
     }
 }
