--- a/net/minecraft/client/renderer/entity/ItemEntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/ItemEntityRenderer.java
@@ -49,7 +_,7 @@
             float bob = Mth.sin(state.ageInTicks / 10.0F + state.bobOffset) * 0.1F + 0.1F;
             poseStack.translate(0.0F, bob + minOffsetY, 0.0F);
             float spin = ItemEntity.getSpin(state.ageInTicks, state.bobOffset);
-            poseStack.rotate(Axis.YP, spin);
+            poseStack.mulPose(Axis.YP.rotation(spin));
             submitMultipleFromCount(poseStack, submitNodeCollector, state.lightCoords, state, this.random, boundingBox);
             poseStack.popPose();
             super.submit(state, poseStack, submitNodeCollector, camera);
