--- a/net/minecraft/client/renderer/entity/LlamaSpitRenderer.java
+++ b/net/minecraft/client/renderer/entity/LlamaSpitRenderer.java
@@ -26,8 +_,8 @@
     public void submit(LlamaSpitRenderState state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
         poseStack.pushPose();
         poseStack.translate(0.0F, 0.15F, 0.0F);
-        poseStack.rotateDegrees(Axis.YP, state.yRot - 90.0F);
-        poseStack.rotateDegrees(Axis.ZP, state.xRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(state.yRot - 90.0F));
+        poseStack.mulPose(Axis.ZP.rotationDegrees(state.xRot));
         submitNodeCollector.submitModel(this.model, state, poseStack, LLAMA_SPIT_LOCATION, state.lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
         poseStack.popPose();
         super.submit(state, poseStack, submitNodeCollector, camera);
