--- a/net/minecraft/client/renderer/entity/OminousItemSpawnerRenderer.java
+++ b/net/minecraft/client/renderer/entity/OminousItemSpawnerRenderer.java
@@ -44,7 +_,7 @@
             }
 
             float currentSpin = Mth.wrapDegrees(state.ageInTicks * 40.0F);
-            poseStack.rotateDegrees(Axis.YP, currentSpin);
+            poseStack.mulPose(Axis.YP.rotationDegrees(currentSpin));
             ItemEntityRenderer.submitMultipleFromCount(poseStack, submitNodeCollector, 15728880, state, this.random);
             poseStack.popPose();
         }
