--- a/net/minecraft/client/renderer/entity/PaintingRenderer.java
+++ b/net/minecraft/client/renderer/entity/PaintingRenderer.java
@@ -37,7 +_,7 @@
         PaintingVariant variant = state.variant;
         if (variant != null) {
             poseStack.pushPose();
-            poseStack.rotateDegrees(Axis.YP, 180 - state.direction.get2DDataValue() * 90);
+            poseStack.mulPose(Axis.YP.rotationDegrees(180 - state.direction.get2DDataValue() * 90));
             TextureAtlasSprite frontSprite = this.paintingsAtlas.getSprite(variant.assetId());
             TextureAtlasSprite backSprite = this.paintingsAtlas.getSprite(BACK_SPRITE_LOCATION);
             this.renderPainting(
