--- a/net/minecraft/client/renderer/entity/PandaRenderer.java
+++ b/net/minecraft/client/renderer/entity/PandaRenderer.java
@@ -98,39 +_,39 @@
                 float nextAngle = 90.0F * nextRollPos / 7.0F;
                 float angle = this.getAngle(thisAngle, nextAngle, nextRollPos, rollTransitionTime, 8.0F);
                 poseStack.translate(0.0F, (y + 0.2F) * (angle / 90.0F), 0.0F);
-                poseStack.rotateDegrees(Axis.XP, -angle);
+                poseStack.mulPose(Axis.XP.rotationDegrees(-angle));
             } else if (rollPos < 16.0F) {
                 float internalRollCounter = (rollPos - 8.0F) / 7.0F;
                 float thisAngle = 90.0F + 90.0F * internalRollCounter;
                 float nextAngle = 90.0F + 90.0F * (nextRollPos - 8.0F) / 7.0F;
                 float angle = this.getAngle(thisAngle, nextAngle, nextRollPos, rollTransitionTime, 16.0F);
                 poseStack.translate(0.0F, y + 0.2F + (y - 0.2F) * (angle - 90.0F) / 90.0F, 0.0F);
-                poseStack.rotateDegrees(Axis.XP, -angle);
+                poseStack.mulPose(Axis.XP.rotationDegrees(-angle));
             } else if (rollPos < 24.0F) {
                 float internalRollCounter = (rollPos - 16.0F) / 7.0F;
                 float thisAngle = 180.0F + 90.0F * internalRollCounter;
                 float nextAngle = 180.0F + 90.0F * (nextRollPos - 16.0F) / 7.0F;
                 float angle = this.getAngle(thisAngle, nextAngle, nextRollPos, rollTransitionTime, 24.0F);
                 poseStack.translate(0.0F, y + y * (270.0F - angle) / 90.0F, 0.0F);
-                poseStack.rotateDegrees(Axis.XP, -angle);
+                poseStack.mulPose(Axis.XP.rotationDegrees(-angle));
             } else if (rollPos < 32) {
                 float internalRollCounter = (rollPos - 24.0F) / 7.0F;
                 float thisAngle = 270.0F + 90.0F * internalRollCounter;
                 float nextAngle = 270.0F + 90.0F * (nextRollPos - 24.0F) / 7.0F;
                 float angle = this.getAngle(thisAngle, nextAngle, nextRollPos, rollTransitionTime, 32.0F);
                 poseStack.translate(0.0F, y * ((360.0F - angle) / 90.0F), 0.0F);
-                poseStack.rotateDegrees(Axis.XP, -angle);
+                poseStack.mulPose(Axis.XP.rotationDegrees(-angle));
             }
         }
 
         float sitAmount = state.sitAmount;
         if (sitAmount > 0.0F) {
             poseStack.translate(0.0F, 0.8F * sitAmount, 0.0F);
-            poseStack.rotateDegrees(Axis.XP, Mth.lerp(sitAmount, state.xRot, state.xRot + 90.0F));
+            poseStack.mulPose(Axis.XP.rotationDegrees(Mth.lerp(sitAmount, state.xRot, state.xRot + 90.0F)));
             poseStack.translate(0.0F, -1.0F * sitAmount, 0.0F);
             if (state.isScared) {
                 float shakeRot = (float)(Math.cos(state.ageInTicks * 1.25F) * Math.PI * 0.05F);
-                poseStack.rotateDegrees(Axis.YP, shakeRot);
+                poseStack.mulPose(Axis.YP.rotationDegrees(shakeRot));
                 if (state.isBaby) {
                     poseStack.translate(0.0F, 0.8F, 0.55F);
                 }
@@ -141,7 +_,7 @@
         if (lieOnBackAmount > 0.0F) {
             float y = state.isBaby ? 0.5F : 1.3F;
             poseStack.translate(0.0F, y * lieOnBackAmount, 0.0F);
-            poseStack.rotateDegrees(Axis.XP, Mth.lerp(lieOnBackAmount, state.xRot, state.xRot + 180.0F));
+            poseStack.mulPose(Axis.XP.rotationDegrees(Mth.lerp(lieOnBackAmount, state.xRot, state.xRot + 180.0F)));
         }
     }
 
