--- a/net/minecraft/client/renderer/entity/PhantomRenderer.java
+++ b/net/minecraft/client/renderer/entity/PhantomRenderer.java
@@ -42,6 +_,6 @@
 
     protected void setupRotations(PhantomRenderState state, PoseStack poseStack, float bodyRot, float entityScale) {
         super.setupRotations(state, poseStack, bodyRot, entityScale);
-        poseStack.rotateDegrees(Axis.XP, state.xRot);
+        poseStack.mulPose(Axis.XP.rotationDegrees(state.xRot));
     }
 }
