--- a/net/minecraft/client/renderer/entity/SalmonRenderer.java
+++ b/net/minecraft/client/renderer/entity/SalmonRenderer.java
@@ -50,10 +_,10 @@
         }
 
         float bodyZRot = amplitudeMultiplier * 4.3F * Mth.sin(angleMultiplier * 0.6F * state.ageInTicks);
-        poseStack.rotateDegrees(Axis.YP, bodyZRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(bodyZRot));
         if (!state.isInWater) {
             poseStack.translate(0.2F, 0.1F, 0.0F);
-            poseStack.rotateDegrees(Axis.ZP, 90.0F);
+            poseStack.mulPose(Axis.ZP.rotationDegrees(90.0F));
         }
     }
 
