--- a/net/minecraft/client/renderer/entity/ShulkerRenderer.java
+++ b/net/minecraft/client/renderer/entity/ShulkerRenderer.java
@@ -33,8 +_,8 @@
         return state.renderOffset;
     }
 
-    public boolean shouldRender(Shulker entity, Frustum culler, double camX, double camY, double camZ, float partialTicks) {
-        if (super.shouldRender(entity, culler, camX, camY, camZ, partialTicks)) {
+    public boolean shouldRender(Shulker entity, Frustum culler, double camX, double camY, double camZ) {
+        if (super.shouldRender(entity, culler, camX, camY, camZ)) {
             return true;
         }
 
