--- a/net/minecraft/client/renderer/entity/SnifferRenderer.java
+++ b/net/minecraft/client/renderer/entity/SnifferRenderer.java
@@ -36,7 +_,7 @@
         state.scentingAnimationState.copyFrom(entity.scentingAnimationState);
     }
 
-    protected AABB getBoundingBoxForCulling(Sniffer entity, float partialTicks) {
-        return super.getBoundingBoxForCulling(entity, partialTicks).inflate(0.6F);
+    protected AABB getBoundingBoxForCulling(Sniffer entity) {
+        return super.getBoundingBoxForCulling(entity).inflate(0.6F);
     }
 }
