--- a/net/minecraft/client/renderer/entity/SquidRenderer.java
+++ b/net/minecraft/client/renderer/entity/SquidRenderer.java
@@ -36,9 +_,9 @@
 
     protected void setupRotations(SquidRenderState state, PoseStack poseStack, float bodyRot, float entityScale) {
         poseStack.translate(0.0F, state.isBaby ? 0.25F : 0.5F, 0.0F);
-        poseStack.rotateDegrees(Axis.YP, 180.0F - bodyRot);
-        poseStack.rotateDegrees(Axis.XP, state.xBodyRot);
-        poseStack.rotateDegrees(Axis.YP, state.zBodyRot);
+        poseStack.mulPose(Axis.YP.rotationDegrees(180.0F - bodyRot));
+        poseStack.mulPose(Axis.XP.rotationDegrees(state.xBodyRot));
+        poseStack.mulPose(Axis.YP.rotationDegrees(state.zBodyRot));
         poseStack.translate(0.0F, state.isBaby ? -0.6F : -1.2F, 0.0F);
     }
 }
