--- a/net/minecraft/client/renderer/entity/ThrownTridentRenderer.java
+++ b/net/minecraft/client/renderer/entity/ThrownTridentRenderer.java
@@ -28,8 +_,8 @@
 
     public void submit(ThrownTridentRenderState state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
         poseStack.pushPose();
-        poseStack.rotateDegrees(Axis.YP, state.yRot - 90.0F);
-        poseStack.rotateDegrees(Axis.ZP, state.xRot + 90.0F);
+        poseStack.mulPose(Axis.YP.rotationDegrees(state.yRot - 90.0F));
+        poseStack.mulPose(Axis.ZP.rotationDegrees(state.xRot + 90.0F));
         if (state.isFoil) {
             submitNodeCollector.submitModel(
                 this.model,
@@ -50,8 +_,8 @@
         super.submit(state, poseStack, submitNodeCollector, camera);
     }
 
-    protected AABB getBoundingBoxForCulling(ThrownTrident entity, float partialTicks) {
-        return super.getBoundingBoxForCulling(entity, partialTicks).inflate(1.5);
+    protected AABB getBoundingBoxForCulling(ThrownTrident entity) {
+        return super.getBoundingBoxForCulling(entity).inflate(1.5);
     }
 
     public ThrownTridentRenderState createRenderState() {
