--- a/net/minecraft/client/renderer/entity/ThrownItemRenderer.java
+++ b/net/minecraft/client/renderer/entity/ThrownItemRenderer.java
@@ -38,7 +_,7 @@
     public void submit(ThrownItemRenderState state, PoseStack poseStack, SubmitNodeCollector submitNodeCollector, CameraRenderState camera) {
         poseStack.pushPose();
         poseStack.scale(this.scale, this.scale, this.scale);
-        poseStack.rotate(camera.orientation);
+        poseStack.mulPose(camera.orientation);
         state.item.submit(poseStack, submitNodeCollector, state.lightCoords, OverlayTexture.NO_OVERLAY, state.outlineColor);
         poseStack.popPose();
         super.submit(state, poseStack, submitNodeCollector, camera);
