--- a/net/minecraft/client/renderer/entity/TntRenderer.java
+++ b/net/minecraft/client/renderer/entity/TntRenderer.java
@@ -32,9 +_,9 @@
             poseStack.scale(scale, scale, scale);
         }
 
-        poseStack.rotateDegrees(Axis.YP, -90.0F);
+        poseStack.mulPose(Axis.YP.rotationDegrees(-90.0F));
         poseStack.translate(-0.5F, -0.5F, 0.5F);
-        poseStack.rotateDegrees(Axis.YP, 90.0F);
+        poseStack.mulPose(Axis.YP.rotationDegrees(90.0F));
         if (!state.blockState.isEmpty()) {
             TntMinecartRenderer.submitWhiteSolidBlock(state.blockState, poseStack, submitNodeCollector, state.lightCoords, isLit(fuse), state.outlineColor);
         }
