--- a/net/minecraft/client/gui/screens/worldselection/AbstractGameRulesScreen.java
+++ b/net/minecraft/client/gui/screens/worldselection/AbstractGameRulesScreen.java
@@ -50,7 +_,7 @@
     private final Set<AbstractGameRulesScreen.RuleEntry> invalidEntries = Sets.newHashSet();
     private final Consumer<Optional<GameRules>> exitCallback;
     protected final HeaderAndFooterLayout layout;
-    public final GameRules gameRules;
+    protected final GameRules gameRules;
     protected @Nullable EditBox searchBox;
     protected AbstractGameRulesScreen.@Nullable RuleList ruleList;
     protected @Nullable Button doneButton;
@@ -101,12 +_,12 @@
         }
     }
 
-    public void markInvalid(AbstractGameRulesScreen.RuleEntry invalidEntry) {
+    private void markInvalid(AbstractGameRulesScreen.RuleEntry invalidEntry) {
         this.invalidEntries.add(invalidEntry);
         this.updateDoneButton();
     }
 
-    public void clearInvalid(AbstractGameRulesScreen.RuleEntry invalidEntry) {
+    private void clearInvalid(AbstractGameRulesScreen.RuleEntry invalidEntry) {
         this.invalidEntries.remove(invalidEntry);
         this.updateDoneButton();
     }
@@ -189,7 +_,7 @@
     }
 
     @FunctionalInterface
-    public interface EntryFactory<T> {
+    private interface EntryFactory<T> {
         AbstractGameRulesScreen.RuleEntry create(Component name, List<FormattedCharSequence> tooltip, String narration, GameRule<T> gameRule);
     }
 
