--- a/net/minecraft/client/gui/screens/worldselection/EditGameRulesScreen.java
+++ b/net/minecraft/client/gui/screens/worldselection/EditGameRulesScreen.java
@@ -30,6 +_,7 @@
 import net.minecraft.client.resources.language.I18n;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.FormattedCharSequence;
 import net.minecraft.world.level.GameRules;
 import net.neoforged.api.distmarker.Dist;
@@ -107,7 +_,7 @@
             this.checkbox = CycleButton.onOffBuilder(p_101104_.get())
                 .displayOnlyValue()
                 .withCustomNarration(p_170219_ -> p_170219_.createDefaultNarrationMessage().append("\n").append(p_101103_))
-                .create(10, 5, 44, 20, p_101101_, (p_170215_, p_170216_) -> p_101104_.set(p_170216_, null));
+                .create(10, 5, 44, 20, p_101101_, (p_170215_, p_170216_) -> p_101104_.set(p_170216_, (MinecraftServer) null));
             this.children.add(this.checkbox);
         }
 
