--- a/net/minecraft/world/entity/ai/behavior/warden/Digging.java
+++ b/net/minecraft/world/entity/ai/behavior/warden/Digging.java
@@ -35,7 +_,8 @@
 
     protected void stop(ServerLevel p_217543_, E p_217544_, long p_217545_) {
         if (p_217544_.getRemovalReason() == null) {
-            p_217544_.remove(Entity.RemovalReason.DISCARDED);
+            p_217544_.setRemoved$cause(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+            p_217544_.remove(Entity.RemovalReason.DISCARDED); // CraftBukkit - Add bukkit remove cause
         }
     }
 }
