--- a/net/minecraft/world/entity/ai/behavior/warden/Digging.java
+++ b/net/minecraft/world/entity/ai/behavior/warden/Digging.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.monster.warden.Warden;
+import org.bukkit.event.entity.EntityRemoveEvent;
 
 public class Digging<E extends Warden> extends Behavior<E> {
     public Digging(int ticks) {
@@ -35,6 +_,7 @@
 
     protected void stop(ServerLevel level, E body, long timestamp) {
         if (body.getRemovalReason() == null) {
+            body.removeReason(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - Add bukkit remove cause
             body.remove(Entity.RemovalReason.DISCARDED);
         }
     }
