--- a/net/minecraft/world/entity/ai/behavior/warden/Digging.java
+++ b/net/minecraft/world/entity/ai/behavior/warden/Digging.java
@@ -9,6 +_,7 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryStatus;
 import net.minecraft.world.entity.monster.warden.Warden;
+import org.bukkit.event.entity.EntityRemoveEvent;
 
 public class Digging<E extends Warden> extends Behavior<E> {
     public Digging(int ticks) {
@@ -20,7 +_,7 @@
     }
 
     protected boolean checkExtraStartConditions(ServerLevel level, E body) {
-        return body.onGround() || body.isInLiquid();
+        return body.onGround() || body.isInWater() || body.isInLava();
     }
 
     protected void start(ServerLevel level, E body, long timestamp) {
@@ -35,7 +_,7 @@
 
     protected void stop(ServerLevel level, E body, long timestamp) {
         if (body.getRemovalReason() == null) {
-            body.remove(Entity.RemovalReason.DISCARDED);
+            body.remove(Entity.RemovalReason.DISCARDED, EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - Add bukkit remove cause
         }
     }
 }
