--- a/net/minecraft/world/entity/ai/behavior/warden/SonicBoom.java
+++ b/net/minecraft/world/entity/ai/behavior/warden/SonicBoom.java
@@ -79,6 +_,7 @@
                     if (p_340767_.hurt(p_217724_.damageSources().sonicBoom(p_217725_), 10.0F)) {
                         double d1 = 0.5 * (1.0 - p_340767_.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE));
                         double d0 = 2.5 * (1.0 - p_340767_.getAttributeValue(Attributes.KNOCKBACK_RESISTANCE));
+                        p_340767_.pushingEntity = p_217725_; // Paper - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
                         p_340767_.push(vec32.x() * d0, vec32.y() * d1, vec32.z() * d0);
                     }
                 });
