--- a/net/minecraft/world/effect/InfestedMobEffect.java
+++ b/net/minecraft/world/effect/InfestedMobEffect.java
@@ -43,7 +_,11 @@
             Vector3f vector3f = p_341023_.getLookAngle().toVector3f().mul(0.3F).mul(1.0F, 1.5F, 1.0F).rotateY(f1);
             silverfish.moveTo(p_338557_, p_338848_, p_338562_, p_338804_.getRandom().nextFloat() * 360.0F, 0.0F);
             silverfish.setDeltaMovement(new Vec3(vector3f));
+            silverfish.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.POTION_EFFECT);
             p_338804_.addFreshEntity(silverfish);
+            if (!p_338804_.canAddFreshEntity())  {
+                return;
+            }
             silverfish.playSound(SoundEvents.SILVERFISH_HURT);
         }
     }
