--- a/net/minecraft/world/effect/InfestedMobEffect.java
+++ b/net/minecraft/world/effect/InfestedMobEffect.java
@@ -44,7 +_,11 @@
             Vector3f viewDirection = mob.getLookAngle().toVector3f().mul(0.3F).mul(1.0F, 1.5F, 1.0F).rotateY(randomAngle);
             silverfish.snapTo(x, y, z, level.getRandom().nextFloat() * 360.0F, 0.0F);
             silverfish.setDeltaMovement(new Vec3(viewDirection));
+            silverfish.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.POTION_EFFECT);
             level.addFreshEntity(silverfish);
+            if (!level.canAddFreshEntity())  {
+                return;
+            }
             silverfish.playSound(SoundEvents.SILVERFISH_HURT);
         }
     }
