--- a/net/minecraft/world/effect/OozingMobEffect.java
+++ b/net/minecraft/world/effect/OozingMobEffect.java
@@ -48,6 +_,7 @@
         if (slime != null) {
             slime.setSize(2, true);
             slime.moveTo(p_338555_, p_338811_, p_338192_, p_338724_.getRandom().nextFloat() * 360.0F, 0.0F);
+            slime.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.POTION_EFFECT);
             p_338724_.addFreshEntity(slime);
         }
     }
