--- a/net/minecraft/world/effect/SaturationMobEffect.java
+++ b/net/minecraft/world/effect/SaturationMobEffect.java
@@ -12,7 +_,16 @@
     @Override
     public boolean applyEffectTick(ServerLevel level, LivingEntity mob, int amplification) {
         if (mob instanceof Player player) {
-            player.getFoodData().eat(amplification + 1, 1.0F);
+            // CraftBukkit start
+            int oldFoodLevel = player.getFoodData().getFoodLevel();
+            org.bukkit.event.entity.FoodLevelChangeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callFoodLevelChangeEvent(player, amplification + 1 + oldFoodLevel);
+            if (!event.isCancelled()) {
+                if (player.level().purpurConfig.playerBurpWhenFull && event.getFoodLevel() == 20 && oldFoodLevel < 20) player.burpDelay = player.level().purpurConfig.playerBurpDelay; // Purpur - Burp delay
+                player.getFoodData().eat(event.getFoodLevel() - oldFoodLevel, mob.level().purpurConfig.humanSaturationRegenAmount); // Purpur - Config MobEffect by world
+            }
+
+            ((org.bukkit.craftbukkit.entity.CraftPlayer) player.getBukkitEntity()).sendHealthUpdate();
+            // CraftBukkit end
         }
 
         return true;
