--- a/net/minecraft/world/level/levelgen/feature/AbstractHugeMushroomFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/AbstractHugeMushroomFeature.java
@@ -3,7 +_,6 @@
 import com.mojang.serialization.MapCodec;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
-import net.minecraft.core.Holder;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.LevelAccessor;
@@ -16,9 +_,9 @@
 public interface AbstractHugeMushroomFeature extends Feature {
     int MIN_MUSHROOM_HEIGHT = 4;
 
-    Holder<BlockStateProvider> capProvider();
+    BlockStateProvider capProvider();
 
-    Holder<BlockStateProvider> stemProvider();
+    BlockStateProvider stemProvider();
 
     int foliageRadius();
 
@@ -30,7 +_,7 @@
     default void placeTrunk(WorldGenLevel level, RandomSource random, BlockPos origin, int treeHeight, BlockPos.MutableBlockPos blockPos) {
         for (int dy = 0; dy < treeHeight; dy++) {
             blockPos.set(origin).move(Direction.UP, dy);
-            this.placeMushroomBlock(level, blockPos, this.stemProvider().value().getState(level, random, origin));
+            this.placeMushroomBlock(level, blockPos, this.stemProvider().getState(level, random, origin));
         }
     }
 
