--- a/net/minecraft/world/level/levelgen/feature/BlockColumnFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/BlockColumnFeature.java
@@ -6,7 +_,6 @@
 import java.util.List;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
-import net.minecraft.core.Holder;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.valueproviders.IntProvider;
 import net.minecraft.util.valueproviders.IntProviders;
@@ -27,16 +_,12 @@
             .apply(i, BlockColumnFeature::new)
     );
 
-    public static BlockColumnFeature.Layer layer(IntProvider height, Holder<BlockStateProvider> state) {
-        return new BlockColumnFeature.Layer(height, state);
-    }
-
     public static BlockColumnFeature.Layer layer(IntProvider height, BlockStateProvider state) {
-        return layer(height, Holder.direct(state));
+        return new BlockColumnFeature.Layer(height, state);
     }
 
     public static BlockColumnFeature simple(IntProvider height, BlockStateProvider state) {
-        return new BlockColumnFeature(List.of(layer(height, Holder.direct(state))), Direction.UP, BlockPredicate.ONLY_IN_AIR_PREDICATE, false);
+        return new BlockColumnFeature(List.of(layer(height, state)), Direction.UP, BlockPredicate.ONLY_IN_AIR_PREDICATE, false);
     }
 
     @Override
@@ -77,7 +_,7 @@
                 BlockColumnFeature.Layer layer = this.layers.get(i);
 
                 for (int y = 0; y < count; y++) {
-                    level.setBlock(placePos, layer.state().value().getState(level, random, placePos), 2);
+                    level.setBlock(placePos, layer.state().getState(level, random, placePos), 2);
                     placePos.move(this.direction);
                 }
             }
@@ -100,7 +_,7 @@
         }
     }
 
-    public record Layer(IntProvider height, Holder<BlockStateProvider> state) {
+    public record Layer(IntProvider height, BlockStateProvider state) {
         public static final Codec<BlockColumnFeature.Layer> CODEC = RecordCodecBuilder.create(
             i -> i.group(
                     IntProviders.NON_NEGATIVE_CODEC.fieldOf("height").forGetter(BlockColumnFeature.Layer::height),
