--- /dev/null
+++ b/net/minecraft/world/level/levelgen/feature/DesertWellFeature.java
@@ -1,0 +_,112 @@
+package net.minecraft.world.level.levelgen.feature;
+
+import com.mojang.serialization.MapCodec;
+import java.util.List;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.Direction;
+import net.minecraft.util.RandomSource;
+import net.minecraft.util.Util;
+import net.minecraft.world.level.WorldGenLevel;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.entity.BlockEntityTypes;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.predicate.BlockStatePredicate;
+import net.minecraft.world.level.chunk.ChunkGenerator;
+import net.minecraft.world.level.storage.loot.BuiltInLootTables;
+
+public record DesertWellFeature() implements Feature {
+    private static final BlockStatePredicate IS_SAND = BlockStatePredicate.forBlock(Blocks.SAND);
+    private static final BlockState SAND = Blocks.SAND.defaultBlockState();
+    private static final BlockState SAND_SLAB = Blocks.SANDSTONE_SLAB.defaultBlockState();
+    private static final BlockState SANDSTONE = Blocks.SANDSTONE.defaultBlockState();
+    private static final BlockState WATER = Blocks.WATER.defaultBlockState();
+    public static final MapCodec<DesertWellFeature> CODEC = MapCodec.unit(DesertWellFeature::new);
+
+    @Override
+    public MapCodec<DesertWellFeature> codec() {
+        return CODEC;
+    }
+
+    @Override
+    public boolean place(WorldGenLevel level, ChunkGenerator chunkGenerator, RandomSource random, BlockPos origin) {
+        origin = origin.above();
+
+        while (level.isEmptyBlock(origin) && origin.getY() > level.getMinY() + 2) {
+            origin = origin.below();
+        }
+
+        if (!IS_SAND.test(level.getBlockState(origin))) {
+            return false;
+        }
+
+        for (int ox = -2; ox <= 2; ox++) {
+            for (int oz = -2; oz <= 2; oz++) {
+                if (level.isEmptyBlock(origin.offset(ox, -1, oz)) && level.isEmptyBlock(origin.offset(ox, -2, oz))) {
+                    return false;
+                }
+            }
+        }
+
+        for (int oy = -2; oy <= 0; oy++) {
+            for (int ox = -2; ox <= 2; ox++) {
+                for (int oz = -2; oz <= 2; oz++) {
+                    level.setBlock(origin.offset(ox, oy, oz), SANDSTONE, 2);
+                }
+            }
+        }
+
+        level.setBlock(origin, WATER, 2);
+
+        for (Direction direction : Direction.Plane.HORIZONTAL) {
+            level.setBlock(origin.relative(direction), WATER, 2);
+        }
+
+        BlockPos sandCenter = origin.below();
+        level.setBlock(sandCenter, SAND, 2);
+
+        for (Direction direction : Direction.Plane.HORIZONTAL) {
+            level.setBlock(sandCenter.relative(direction), SAND, 2);
+        }
+
+        for (int ox = -2; ox <= 2; ox++) {
+            for (int oz = -2; oz <= 2; oz++) {
+                if (ox == -2 || ox == 2 || oz == -2 || oz == 2) {
+                    level.setBlock(origin.offset(ox, 1, oz), SANDSTONE, 2);
+                }
+            }
+        }
+
+        level.setBlock(origin.offset(2, 1, 0), SAND_SLAB, 2);
+        level.setBlock(origin.offset(-2, 1, 0), SAND_SLAB, 2);
+        level.setBlock(origin.offset(0, 1, 2), SAND_SLAB, 2);
+        level.setBlock(origin.offset(0, 1, -2), SAND_SLAB, 2);
+
+        for (int ox = -1; ox <= 1; ox++) {
+            for (int oz = -1; oz <= 1; oz++) {
+                if (ox == 0 && oz == 0) {
+                    level.setBlock(origin.offset(ox, 4, oz), SANDSTONE, 2);
+                } else {
+                    level.setBlock(origin.offset(ox, 4, oz), SAND_SLAB, 2);
+                }
+            }
+        }
+
+        for (int oy = 1; oy <= 3; oy++) {
+            level.setBlock(origin.offset(-1, oy, -1), SANDSTONE, 2);
+            level.setBlock(origin.offset(-1, oy, 1), SANDSTONE, 2);
+            level.setBlock(origin.offset(1, oy, -1), SANDSTONE, 2);
+            level.setBlock(origin.offset(1, oy, 1), SANDSTONE, 2);
+        }
+
+        BlockPos waterCenter = origin;
+        List<BlockPos> waterPositions = List.of(waterCenter, waterCenter.east(), waterCenter.south(), waterCenter.west(), waterCenter.north());
+        placeSusSand(level, Util.getRandom(waterPositions, random).below(1));
+        placeSusSand(level, Util.getRandom(waterPositions, random).below(2));
+        return true;
+    }
+
+    private static void placeSusSand(WorldGenLevel level, BlockPos pos) {
+        level.setBlockAndUpdate(pos, Blocks.SUSPICIOUS_SAND.defaultBlockState());
+        level.getBlockEntity(pos, BlockEntityTypes.BRUSHABLE_BLOCK).ifPresent(e -> e.setLootTable(BuiltInLootTables.DESERT_WELL_ARCHAEOLOGY, pos.asLong()));
+    }
+}
