--- a/net/minecraft/world/level/levelgen/feature/DiskFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/DiskFeature.java
@@ -4,7 +_,6 @@
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.BlockPos;
-import net.minecraft.core.Holder;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.valueproviders.IntProvider;
 import net.minecraft.util.valueproviders.IntProviders;
@@ -14,7 +_,7 @@
 import net.minecraft.world.level.levelgen.blockpredicates.BlockPredicate;
 import net.minecraft.world.level.levelgen.feature.stateproviders.BlockStateProvider;
 
-public record DiskFeature(Holder<BlockStateProvider> stateProvider, BlockPredicate target, IntProvider radius, int halfHeight) implements Feature {
+public record DiskFeature(BlockStateProvider stateProvider, BlockPredicate target, IntProvider radius, int halfHeight) implements Feature {
     public static final MapCodec<DiskFeature> CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(
                 BlockStateProvider.CODEC.fieldOf("state_provider").forGetter(DiskFeature::stateProvider),
@@ -57,7 +_,7 @@
         for (int y = top; y > bottom; y--) {
             pos.setY(y);
             if (this.target.test(level, pos)) {
-                BlockState state = this.stateProvider.value().getOptionalState(level, random, pos);
+                BlockState state = this.stateProvider.getOptionalState(level, random, pos);
                 if (state != null) {
                     level.setBlock(pos, state, 2);
                     if (!placedAbove) {
