--- a/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
@@ -1,11 +_,14 @@
 package net.minecraft.world.level.levelgen.feature;
 
 import com.mojang.serialization.Codec;
+import java.util.concurrent.atomic.AtomicReference;
 import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;
+import org.bukkit.craftbukkit.util.BlockStateListPopulator;
 
 public class EndPlatformFeature extends Feature<NoneFeatureConfiguration> {
     public EndPlatformFeature(Codec<NoneFeatureConfiguration> p_352966_) {
@@ -18,7 +_,12 @@
         return true;
     }
 
+    private static final AtomicReference<Entity> createEndPlatform$entity = new AtomicReference<>();
+
     public static void createEndPlatform(ServerLevelAccessor p_352905_, BlockPos p_352961_, boolean p_352931_) {
+        BlockStateListPopulator blockList = new BlockStateListPopulator(p_352905_);
+        // CraftBukkit end
+
         BlockPos.MutableBlockPos blockpos$mutableblockpos = p_352961_.mutable();
 
         for (int i = -2; i <= 2; i++) {
@@ -26,15 +_,38 @@
                 for (int k = -1; k < 3; k++) {
                     BlockPos blockpos = blockpos$mutableblockpos.set(p_352961_).move(j, k, i);
                     Block block = k == -1 ? Blocks.OBSIDIAN : Blocks.AIR;
-                    if (!p_352905_.getBlockState(blockpos).is(block)) {
+                    // CraftBukkit start
+                    if (!blockList.getBlockState(blockpos).is(block)) {
                         if (p_352931_) {
-                            p_352905_.destroyBlock(blockpos, true, null);
+                            blockList.destroyBlock(blockpos, true, null);
                         }
 
-                        p_352905_.setBlock(blockpos, block.defaultBlockState(), 3);
+                        blockList.setBlock(blockpos, block.defaultBlockState(), 3);
+                        // CraftBukkit end
                     }
                 }
             }
         }
+
+        // CraftBukkit start
+        if (createEndPlatform$entity.get() == null) {
+            // SPIGOT-7746: Entity will only be null during world generation, which is async, so just generate without event
+            blockList.updateList();
+            return;
+        }
+
+        org.bukkit.World bworld = p_352905_.getLevel().getWorld();
+        org.bukkit.event.world.PortalCreateEvent portalEvent = new org.bukkit.event.world.PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blockList.getList(), bworld, createEndPlatform$entity.getAndSet(null).getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.END_PLATFORM);
+
+        p_352905_.getLevel().getCraftServer().getPluginManager().callEvent(portalEvent);
+        if (!portalEvent.isCancelled()) {
+            blockList.updateList();
+        }
+        // CraftBukkit end
+    }
+
+    public static void createEndPlatform(ServerLevelAccessor p_352905_, BlockPos p_352961_, boolean p_352931_, net.minecraft.world.entity.Entity entity) {
+        createEndPlatform$entity.set(entity);
+        createEndPlatform(p_352905_, p_352961_, p_352931_);
     }
 }
