--- a/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/EndPlatformFeature.java
@@ -7,6 +_,12 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;
 
+// CraftBukkit start
+import java.util.List;
+import org.bukkit.block.BlockState;
+import org.bukkit.event.world.PortalCreateEvent;
+// CraftBukkit end
+
 public class EndPlatformFeature extends Feature<NoneFeatureConfiguration> {
     public EndPlatformFeature(Codec<NoneFeatureConfiguration> codec) {
         super(codec);
@@ -19,6 +_,13 @@
     }
 
     public static void createEndPlatform(ServerLevelAccessor newLevel, BlockPos origin, boolean dropResources) {
+        createEndPlatform(newLevel, origin, dropResources, null);
+        // CraftBukkit start
+    }
+
+    public static void createEndPlatform(ServerLevelAccessor newLevel, BlockPos origin, boolean dropResources, net.minecraft.world.entity.Entity entity) {
+        org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(newLevel);
+        // CraftBukkit end
         BlockPos.MutableBlockPos pos = origin.mutable();
 
         for (int dz = -2; dz <= 2; dz++) {
@@ -26,15 +_,35 @@
                 for (int dy = -1; dy < 3; dy++) {
                     BlockPos blockPos = pos.set(origin).move(dx, dy, dz);
                     Block block = dy == -1 ? Blocks.OBSIDIAN : Blocks.AIR;
-                    if (!newLevel.getBlockState(blockPos).is(block)) {
+                    // CraftBukkit start
+                    if (!blockList.getBlockState(blockPos).is(block)) {
                         if (dropResources) {
-                            newLevel.destroyBlock(blockPos, true, null);
+                            blockList.destroyBlock(blockPos, true, null);
                         }
 
-                        newLevel.setBlock(blockPos, block.defaultBlockState(), 3);
+                        blockList.setBlock(blockPos, block.defaultBlockState(), 3);
+                        // CraftBukkit end
                     }
                 }
             }
         }
+        // CraftBukkit start
+        // SPIGOT-7746: Entity will only be null during world generation, which is async, so just generate without event
+        if (entity != null) {
+            org.bukkit.World bworld = newLevel.getLevel().getWorld();
+            PortalCreateEvent portalEvent = new PortalCreateEvent((List<BlockState>) (List) blockList.getList(), bworld, entity.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.END_PLATFORM);
+
+            newLevel.getLevel().getCraftServer().getPluginManager().callEvent(portalEvent);
+            if (portalEvent.isCancelled()) {
+                return;
+            }
+        }
+
+        // SPIGOT-7856: End platform not dropping items after replacing blocks
+        if (dropResources) {
+            blockList.getList().forEach((state) -> newLevel.destroyBlock(state.getPosition(), true, null));
+        }
+        blockList.updateList();
+        // CraftBukkit end
     }
 }
