--- a/net/minecraft/world/level/levelgen/feature/EndSpikeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/EndSpikeFeature.java
@@ -117,8 +_,9 @@
         EndCrystal endCrystal = EntityTypes.END_CRYSTAL.create(level.getLevel(), EntitySpawnReason.STRUCTURE);
         if (endCrystal != null) {
             endCrystal.setBeamTarget(this.crystalBeamTarget.orElse(null));
-            endCrystal.setPermanentlyInvulnerable(this.crystalInvulnerable);
+            endCrystal.setInvulnerable(this.crystalInvulnerable);
             endCrystal.snapTo(spike.getCenterX() + 0.5, spike.getHeight() + 1, spike.getCenterZ() + 0.5, random.nextFloat() * 360.0F, 0.0F);
+            endCrystal.generatedByDragonFight = true; // Paper - Fix invulnerable end crystals
             level.addFreshEntity(endCrystal);
             BlockPos crystalPos = endCrystal.blockPosition();
             this.setBlock(level, crystalPos.below(), Blocks.BEDROCK.defaultBlockState());
