--- a/net/minecraft/world/level/levelgen/feature/FeatureTypes.java
+++ b/net/minecraft/world/level/levelgen/feature/FeatureTypes.java
@@ -15,6 +_,7 @@
         Registry.register(registry, "coral_claw", CoralClawFeature.CODEC);
         Registry.register(registry, "coral_tree", CoralTreeFeature.CODEC);
         Registry.register(registry, "delta_feature", DeltaFeature.CODEC);
+        Registry.register(registry, "desert_well", DesertWellFeature.CODEC);
         Registry.register(registry, "disk", DiskFeature.CODEC);
         Registry.register(registry, "end_gateway", EndGatewayFeature.CODEC);
         Registry.register(registry, "end_island", EndIslandFeature.CODEC);
