--- a/net/minecraft/world/level/levelgen/feature/FossilFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/FossilFeature.java
@@ -59,7 +_,7 @@
     public boolean place(WorldGenLevel level, ChunkGenerator chunkGenerator, RandomSource random, BlockPos origin) {
         Rotation rotation = Rotation.getRandom(random);
         int fossilIndex = random.nextInt(this.fossilStructures().size());
-        StructureTemplateManager structureTemplateManager = level.getLevel().getServer().getStructureTemplateManager();
+        StructureTemplateManager structureTemplateManager = level.getLevel().getServer().getStructureManager();
         StructureTemplate fossilBase = structureTemplateManager.getOrCreate(this.fossilStructures().get(fossilIndex));
         StructureTemplate fossilOverlay = structureTemplateManager.getOrCreate(this.overlayStructures().get(fossilIndex));
         ChunkPos chunkPos = ChunkPos.containing(origin);
