--- a/net/minecraft/world/level/levelgen/feature/GeodeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/GeodeFeature.java
@@ -28,7 +_,6 @@
 import net.minecraft.world.level.levelgen.GeodeLayerSettings;
 import net.minecraft.world.level.levelgen.LegacyRandomSource;
 import net.minecraft.world.level.levelgen.WorldgenRandom;
-import net.minecraft.world.level.levelgen.synth.Noise;
 import net.minecraft.world.level.levelgen.synth.NormalNoise;
 import net.minecraft.world.level.material.FluidState;
 
@@ -47,7 +_,6 @@
     double noiseMultiplier,
     int invalidBlocksThreshold
 ) implements Feature {
-    private static final NormalNoise NOISE_PARAMETERS = NormalNoise.createParity(-4, 1.0);
     public static final Codec<Double> CHANCE_RANGE = Codec.doubleRange(0.0, 1.0);
     public static final MapCodec<GeodeFeature> CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(
@@ -78,7 +_,8 @@
     public boolean place(WorldGenLevel level, ChunkGenerator chunkGenerator, RandomSource random, BlockPos origin) {
         List<Pair<BlockPos, Integer>> points = Lists.newLinkedList();
         int numPoints = this.distributionPoints.sample(random);
-        Noise noise = NOISE_PARAMETERS.create(new WorldgenRandom(new LegacyRandomSource(level.getSeed())));
+        WorldgenRandom random1 = new WorldgenRandom(new LegacyRandomSource(level.getSeed()));
+        NormalNoise noise = NormalNoise.create(random1, -4, 1.0);
         List<BlockPos> crackPoints = Lists.newLinkedList();
         double crackSizeAdjustment = (double)numPoints / this.outerWallDistance.maxInclusive();
         double innerAir = 1.0 / Math.sqrt(this.layerSettings.filling);
@@ -133,7 +_,7 @@
         for (BlockPos pointInside : BlockPos.betweenClosed(
             origin.offset(this.minGenOffset, this.minGenOffset, this.minGenOffset), origin.offset(this.maxGenOffset, this.maxGenOffset, this.maxGenOffset)
         )) {
-            double noiseOffset = noise.get(pointInside.getX(), pointInside.getY(), pointInside.getZ()) * this.noiseMultiplier;
+            double noiseOffset = noise.getValue(pointInside.getX(), pointInside.getY(), pointInside.getZ()) * this.noiseMultiplier;
             double distSumShell = 0.0;
 
             for (Pair<BlockPos, Integer> point : points) {
@@ -142,7 +_,7 @@
 
             if (!(distSumShell < outerCrust)) {
                 if (distSumShell >= innerAir) {
-                    this.safeSetBlock(level, pointInside, this.blockSettings.fillingProvider().value().getState(level, random, pointInside), canReplace);
+                    this.safeSetBlock(level, pointInside, this.blockSettings.fillingProvider().getState(level, random, pointInside), canReplace);
                 } else {
                     double distSumCrack = 0.0;
 
@@ -164,21 +_,19 @@
                         boolean useAlternateLayer = random.nextFloat() < this.useAlternateLayer0Chance;
                         if (useAlternateLayer) {
                             this.safeSetBlock(
-                                level, pointInside, this.blockSettings.alternateInnerLayerProvider().value().getState(level, random, pointInside), canReplace
+                                level, pointInside, this.blockSettings.alternateInnerLayerProvider().getState(level, random, pointInside), canReplace
                             );
                         } else {
-                            this.safeSetBlock(
-                                level, pointInside, this.blockSettings.innerLayerProvider().value().getState(level, random, pointInside), canReplace
-                            );
+                            this.safeSetBlock(level, pointInside, this.blockSettings.innerLayerProvider().getState(level, random, pointInside), canReplace);
                         }
 
                         if ((!this.placementsRequireLayer0Alternate || useAlternateLayer) && random.nextFloat() < this.usePotentialPlacementsChance) {
                             potentialCrystalPlacements.add(pointInside.immutable());
                         }
                     } else if (distSumShell >= innerCrust) {
-                        this.safeSetBlock(level, pointInside, this.blockSettings.middleLayerProvider().value().getState(level, random, pointInside), canReplace);
+                        this.safeSetBlock(level, pointInside, this.blockSettings.middleLayerProvider().getState(level, random, pointInside), canReplace);
                     } else if (distSumShell >= outerCrust) {
-                        this.safeSetBlock(level, pointInside, this.blockSettings.outerLayerProvider().value().getState(level, random, pointInside), canReplace);
+                        this.safeSetBlock(level, pointInside, this.blockSettings.outerLayerProvider().getState(level, random, pointInside), canReplace);
                     }
                 }
             }
