--- a/net/minecraft/world/level/levelgen/feature/LakeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/LakeFeature.java
@@ -3,7 +_,6 @@
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.BlockPos;
-import net.minecraft.core.Holder;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.WorldGenLevel;
@@ -15,8 +_,8 @@
 
 @Deprecated
 public record LakeFeature(
-    Holder<BlockStateProvider> fluid,
-    Holder<BlockStateProvider> barrier,
+    BlockStateProvider fluid,
+    BlockStateProvider barrier,
     BlockPredicate canPlaceFeature,
     BlockPredicate canReplaceWithAirOrFluid,
     BlockPredicate canReplaceWithBarrier
@@ -71,7 +_,7 @@
             }
         }
 
-        BlockState fluid = this.fluid.value().getState(level, random, origin);
+        BlockState fluid = this.fluid.getState(level, random, origin);
 
         for (int xx = 0; xx < 16; xx++) {
             for (int zz = 0; zz < 16; zz++) {
@@ -122,7 +_,7 @@
             }
         }
 
-        BlockState barrier = this.barrier.value().getState(level, random, origin);
+        BlockState barrier = this.barrier.getState(level, random, origin);
         if (!barrier.isAir()) {
             for (int xx = 0; xx < 16; xx++) {
                 for (int zz = 0; zz < 16; zz++) {
