--- a/net/minecraft/world/level/levelgen/feature/MonsterRoomFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/MonsterRoomFeature.java
@@ -121,7 +_,7 @@
             if (worldgenlevel.getBlockEntity(blockpos) instanceof SpawnerBlockEntity spawnerblockentity) {
                 spawnerblockentity.setEntityId(this.randomEntityId(randomsource), randomsource);
             } else {
-                LOGGER.error("Failed to fetch mob spawner entity at ({}, {}, {})", blockpos.getX(), blockpos.getY(), blockpos.getZ());
+                LOGGER.debug("Failed to fetch mob spawner entity at ({}, {}, {})", blockpos.getX(), blockpos.getY(), blockpos.getZ());
             }
 
             return true;
@@ -131,6 +_,6 @@
     }
 
     private EntityType<?> randomEntityId(RandomSource p_225154_) {
-        return Util.getRandom(MOBS, p_225154_);
+        return net.neoforged.neoforge.common.MonsterRoomHooks.getRandomMonsterRoomMob(p_225154_);
     }
 }
