--- a/net/minecraft/world/level/levelgen/feature/RootSystemFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/RootSystemFeature.java
@@ -7,8 +_,8 @@
 import net.minecraft.core.Direction;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.WorldGenLevel;
@@ -27,12 +_,12 @@
     int maxLevelDeviation,
     int rootRadius,
     HolderSet<Block> rootReplaceable,
-    Holder<BlockStateProvider> rootStateProvider,
+    BlockStateProvider rootStateProvider,
     int rootPlacementAttempts,
     int rootColumnMaxHeight,
     int hangingRootRadius,
     int hangingRootsVerticalSpan,
-    Holder<BlockStateProvider> hangingRootStateProvider,
+    BlockStateProvider hangingRootStateProvider,
     int hangingRootPlacementAttempts,
     int allowedVerticalWaterForTree,
     BlockPredicate allowedTreePosition
@@ -44,7 +_,7 @@
                 Codec.intRange(0, 16).fieldOf("level_test_distance").forGetter(RootSystemFeature::levelTestDistance),
                 Codec.intRange(0, 64).fieldOf("max_level_deviation").forGetter(RootSystemFeature::maxLevelDeviation),
                 Codec.intRange(1, 64).fieldOf("root_radius").forGetter(RootSystemFeature::rootRadius),
-                RegistryCodecs.holderSet(Registries.BLOCK).fieldOf("root_replaceable").forGetter(RootSystemFeature::rootReplaceable),
+                RegistryCodecs.homogeneousList(Registries.BLOCK).fieldOf("root_replaceable").forGetter(RootSystemFeature::rootReplaceable),
                 BlockStateProvider.CODEC.fieldOf("root_state_provider").forGetter(RootSystemFeature::rootStateProvider),
                 Codec.intRange(1, 256).fieldOf("root_placement_attempts").forGetter(RootSystemFeature::rootPlacementAttempts),
                 Codec.intRange(1, 4096).fieldOf("root_column_max_height").forGetter(RootSystemFeature::rootColumnMaxHeight),
@@ -157,7 +_,7 @@
                 random.nextInt(this.rootRadius) - random.nextInt(this.rootRadius)
             );
             if (level.getBlockState(workingPos).is(this.rootReplaceable)) {
-                level.setBlock(workingPos, this.rootStateProvider.value().getState(level, random, workingPos), 2);
+                level.setBlock(workingPos, this.rootStateProvider.getState(level, random, workingPos), 2);
             }
 
             workingPos.setX(originX);
@@ -174,7 +_,7 @@
                 random.nextInt(this.hangingRootRadius) - random.nextInt(this.hangingRootRadius)
             );
             if (level.isEmptyBlock(workingPos)) {
-                BlockState targetState = this.hangingRootStateProvider.value().getState(level, random, workingPos);
+                BlockState targetState = this.hangingRootStateProvider.getState(level, random, workingPos);
                 if (targetState.canSurvive(level, workingPos) && level.getBlockState(workingPos.above()).isFaceSturdy(level, workingPos, Direction.DOWN)) {
                     level.setBlock(workingPos, targetState, 2);
                 }
