--- a/net/minecraft/world/level/levelgen/feature/SpringFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/SpringFeature.java
@@ -5,8 +_,8 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.WorldGenLevel;
 import net.minecraft.world.level.block.Block;
@@ -21,7 +_,7 @@
                 Codec.BOOL.optionalFieldOf("requires_block_below", true).forGetter(SpringFeature::requiresBlockBelow),
                 Codec.INT.optionalFieldOf("rock_count", 4).forGetter(SpringFeature::rockCount),
                 Codec.INT.optionalFieldOf("hole_count", 1).forGetter(SpringFeature::holeCount),
-                RegistryCodecs.holderSet(Registries.BLOCK).fieldOf("valid_blocks").forGetter(SpringFeature::validBlocks)
+                RegistryCodecs.homogeneousList(Registries.BLOCK).fieldOf("valid_blocks").forGetter(SpringFeature::validBlocks)
             )
             .apply(i, SpringFeature::new)
     );
