--- a/net/minecraft/world/level/levelgen/feature/TemplateFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/TemplateFeature.java
@@ -4,10 +_,8 @@
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import java.util.List;
-import java.util.Optional;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
-import net.minecraft.core.Holder;
 import net.minecraft.core.Vec3i;
 import net.minecraft.resources.Identifier;
 import net.minecraft.util.RandomSource;
@@ -17,33 +_,15 @@
 import net.minecraft.world.level.block.Rotation;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructurePlaceSettings;
-import net.minecraft.world.level.levelgen.structure.templatesystem.StructureProcessor;
-import net.minecraft.world.level.levelgen.structure.templatesystem.StructureProcessorList;
-import net.minecraft.world.level.levelgen.structure.templatesystem.StructureProcessorType;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplate;
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
 
-public record TemplateFeature(WeightedList<TemplateFeature.TemplateEntry> templates, Optional<Holder<StructureProcessorList>> processors) implements Feature {
+public record TemplateFeature(WeightedList<TemplateFeature.TemplateEntry> templates) implements Feature {
     public static final MapCodec<TemplateFeature> CODEC = RecordCodecBuilder.mapCodec(
-        i -> i.group(
-                WeightedList.codec(TemplateFeature.TemplateEntry.CODEC).fieldOf("templates").forGetter(TemplateFeature::templates),
-                StructureProcessorType.LIST_CODEC.optionalFieldOf("processors").forGetter(TemplateFeature::processors)
-            )
+        i -> i.group(WeightedList.codec(TemplateFeature.TemplateEntry.CODEC).fieldOf("templates").forGetter(TemplateFeature::templates))
             .apply(i, TemplateFeature::new)
     );
 
-    public TemplateFeature(WeightedList<TemplateFeature.TemplateEntry> templates) {
-        this(templates, Optional.empty());
-    }
-
-    public static TemplateFeature simple(Identifier id, Holder<StructureProcessorList> processors) {
-        return new TemplateFeature(WeightedList.of(TemplateFeature.TemplateEntry.of(id)), Optional.of(processors));
-    }
-
-    public static TemplateFeature simple(Identifier id) {
-        return new TemplateFeature(WeightedList.of(TemplateFeature.TemplateEntry.of(id)), Optional.empty());
-    }
-
     @Override
     public MapCodec<TemplateFeature> codec() {
         return CODEC;
@@ -53,18 +_,12 @@
     public boolean place(WorldGenLevel level, ChunkGenerator chunkGenerator, RandomSource random, BlockPos origin) {
         TemplateFeature.TemplateEntry templateEntry = this.templates.getRandomOrThrow(random);
         Rotation rotation = Util.getRandom(templateEntry.rotations(), random);
-        StructureTemplateManager structureTemplateManager = level.getLevel().getServer().getStructureTemplateManager();
+        StructureTemplateManager structureTemplateManager = level.getLevel().getServer().getStructureManager();
         StructureTemplate template = structureTemplateManager.getOrCreate(templateEntry.template());
         Vec3i offsetX = this.getRotatedOffset(rotation, Direction.Axis.X, template);
         Vec3i offsetZ = this.getRotatedOffset(rotation, Direction.Axis.Z, template);
         BlockPos pos = origin.offset(offsetX).offset(offsetZ);
         StructurePlaceSettings settings = new StructurePlaceSettings().setRotation(rotation).setRandom(random);
-        if (this.processors.isPresent()) {
-            for (StructureProcessor processor : this.processors.get().value().list()) {
-                settings.addProcessor(processor);
-            }
-        }
-
         return template.placeInWorld(level, pos, pos, settings, random, 3);
     }
 
