--- a/net/minecraft/world/level/levelgen/feature/WaterloggedVegetationPatchFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/WaterloggedVegetationPatchFeature.java
@@ -25,7 +_,7 @@
 
     public WaterloggedVegetationPatchFeature(
         HolderSet<Block> replaceable,
-        Holder<BlockStateProvider> groundState,
+        BlockStateProvider groundState,
         Holder<PlacedFeature> vegetationFeature,
         CaveSurface surface,
         IntProvider depth,
