--- a/net/minecraft/world/entity/animal/feline/Cat.java
+++ b/net/minecraft/world/entity/animal/feline/Cat.java
@@ -473,7 +_,7 @@
     }
 
     private void tryToTame(Player player) {
-        if (this.random.nextInt(3) == 0) {
+        if (this.random.nextInt(3) == 0 && !net.neoforged.neoforge.event.EventHooks.onAnimalTame(this, player) && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTameEvent(this, player).isCancelled()) { // CraftBukkit
             this.tame(player);
             this.setOrderedToSit(true);
             this.level().broadcastEntityEvent(this, (byte)7);
@@ -601,15 +_,24 @@
                 .dropFromGiftLootTable(
                     getServerLevel(this.cat),
                     BuiltInLootTables.CAT_MORNING_GIFT,
-                    (level, itemStack) -> level.addFreshEntity(
-                        new ItemEntity(
-                            level,
-                            (double)catPos.getX() - Mth.sin(this.cat.yBodyRot * (float) (Math.PI / 180.0)),
-                            catPos.getY(),
-                            (double)catPos.getZ() + Mth.cos(this.cat.yBodyRot * (float) (Math.PI / 180.0)),
-                            itemStack
-                        )
-                    )
+                    (level, itemStack) -> {
+                        // CraftBukkit start
+                        ItemEntity itementity =
+                                new ItemEntity(
+                                        level,
+                                        (double) catPos.getX() - Mth.sin(this.cat.yBodyRot * (float) (Math.PI / 180.0)),
+                                        catPos.getY(),
+                                        (double) catPos.getZ() + Mth.cos(this.cat.yBodyRot * (float) (Math.PI / 180.0)),
+                                        itemStack
+                                );
+                        org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(this.cat.getBukkitEntity(), (org.bukkit.entity.Item) itementity.getBukkitEntity());
+                        itementity.level().getCraftServer().getPluginManager().callEvent(event);
+                        if (event.isCancelled()) {
+                            return;
+                        }
+                        level.addFreshEntity(itementity);
+                        // CraftBukkit end
+                    }
                 );
         }
 
@@ -635,7 +_,7 @@
     }
 
     private static class CatTemptGoal extends TemptGoal {
-        private @Nullable Player selectedPlayer;
+        private @Nullable LivingEntity selectedPlayer;
         private final Cat cat;
 
         public CatTemptGoal(Cat mob, double speedModifier, Predicate<ItemStack> items, boolean canScare) {
