--- a/com/mojang/renderpearl/api/buffers/GpuBufferSlice.java
+++ b/com/mojang/renderpearl/api/buffers/GpuBufferSlice.java
@@ -1,6 +_,5 @@
 package com.mojang.renderpearl.api.buffers;
 
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.nio.ByteBuffer;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
@@ -29,7 +_,7 @@
         return this.buffer.map(this.offset, this.length, read, write);
     }
 
-    public record MappedView(GpuBufferSlice slice, ByteBuffer data, Runnable onClose) implements UncheckedAutoCloseable {
+    public record MappedView(GpuBufferSlice slice, ByteBuffer data, Runnable onClose) implements AutoCloseable {
         @Override
         public void close() {
             this.onClose.run();
