--- a/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
+++ b/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
@@ -54,7 +_,7 @@
 import net.minecraft.world.entity.monster.Blaze;
 import net.minecraft.world.entity.monster.Creeper;
 import net.minecraft.world.entity.monster.ElderGuardian;
-import net.minecraft.world.entity.monster.Enderman;
+import net.minecraft.world.entity.monster.EnderMan;
 import net.minecraft.world.entity.monster.Endermite;
 import net.minecraft.world.entity.monster.Ghast;
 import net.minecraft.world.entity.monster.Giant;
@@ -116,7 +_,7 @@
         .put(EntityTypes.DONKEY, AbstractChestedHorse.createBaseChestedHorseAttributes().build())
         .put(EntityTypes.DROWNED, Drowned.createAttributes().build())
         .put(EntityTypes.ELDER_GUARDIAN, ElderGuardian.createAttributes().build())
-        .put(EntityTypes.ENDERMAN, Enderman.createAttributes().build())
+        .put(EntityTypes.ENDERMAN, EnderMan.createAttributes().build())
         .put(EntityTypes.ENDERMITE, Endermite.createAttributes().build())
         .put(EntityTypes.ENDER_DRAGON, EnderDragon.createAttributes().build())
         .put(EntityTypes.EVOKER, Evoker.createAttributes().build())
@@ -138,14 +_,14 @@
         .put(EntityTypes.MAGMA_CUBE, MagmaCube.createAttributes().build())
         .put(EntityTypes.SULFUR_CUBE, SulfurCube.createSulfurCubeAttributes().build())
         .put(EntityTypes.MANNEQUIN, LivingEntity.createLivingAttributes().build())
-        .put(EntityTypes.MOOSHROOM, Cow.createAttributes().build())
+        .put(EntityTypes.MOOSHROOM, net.minecraft.world.entity.animal.cow.AbstractCow.createAttributes().build()) // Purpur - Cows naturally aggressive to players chance
         .put(EntityTypes.MULE, AbstractChestedHorse.createBaseChestedHorseAttributes().build())
         .put(EntityTypes.NAUTILUS, Nautilus.createAttributes().build())
         .put(EntityTypes.OCELOT, Ocelot.createAttributes().build())
         .put(EntityTypes.PANDA, Panda.createAttributes().build())
         .put(EntityTypes.PARCHED, Parched.createAttributes().build())
         .put(EntityTypes.PARROT, Parrot.createAttributes().build())
-        .put(EntityTypes.PHANTOM, Monster.createMonsterAttributes().build())
+        .put(EntityTypes.PHANTOM, net.minecraft.world.entity.monster.Phantom.createAttributes().build()) // Purpur - Ridables
         .put(EntityTypes.PIG, Pig.createAttributes().build())
         .put(EntityTypes.PIGLIN, Piglin.createAttributes().build())
         .put(EntityTypes.PIGLIN_BRUTE, PiglinBrute.createAttributes().build())
@@ -176,7 +_,7 @@
         .put(EntityTypes.VILLAGER, Villager.createAttributes().build())
         .put(EntityTypes.VINDICATOR, Vindicator.createAttributes().build())
         .put(EntityTypes.WARDEN, Warden.createAttributes().build())
-        .put(EntityTypes.WANDERING_TRADER, Mob.createMobAttributes().build())
+        .put(EntityTypes.WANDERING_TRADER, net.minecraft.world.entity.npc.wanderingtrader.WanderingTrader.createAttributes().build()) // Purpur - Villagers follow emerald blocks
         .put(EntityTypes.WITCH, Witch.createAttributes().build())
         .put(EntityTypes.WITHER, WitherBoss.createAttributes().build())
         .put(EntityTypes.WITHER_SKELETON, AbstractSkeleton.createAttributes().build())
@@ -190,11 +_,12 @@
         .build();
 
     public static AttributeSupplier getSupplier(EntityType<? extends LivingEntity> type) {
-        return SUPPLIERS.get(type);
+        AttributeSupplier supplier = net.neoforged.neoforge.common.CommonHooks.getAttributesView().get(type);
+        return supplier != null ? supplier : SUPPLIERS.get(type);
     }
 
     public static boolean hasSupplier(EntityType<?> type) {
-        return SUPPLIERS.containsKey(type);
+        return SUPPLIERS.containsKey(type) || net.neoforged.neoforge.common.CommonHooks.getAttributesView().containsKey(type);
     }
 
     public static void validate() {
