--- a/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
+++ b/net/minecraft/world/entity/ai/attributes/DefaultAttributes.java
@@ -145,7 +_,7 @@
         .put(EntityTypes.PANDA, Panda.createAttributes().build())
         .put(EntityTypes.PARCHED, Parched.createAttributes().build())
         .put(EntityTypes.PARROT, Parrot.createAttributes().build())
-        .put(EntityTypes.PHANTOM, Monster.createMonsterAttributes().build())
+        .put(EntityTypes.PHANTOM, net.minecraft.world.entity.monster.Phantom.createAttributes().build()) // Purpur - Ridables
         .put(EntityTypes.PIG, Pig.createAttributes().build())
         .put(EntityTypes.PIGLIN, Piglin.createAttributes().build())
         .put(EntityTypes.PIGLIN_BRUTE, PiglinBrute.createAttributes().build())
@@ -190,11 +_,12 @@
         .build();
 
     public static AttributeSupplier getSupplier(EntityType<? extends LivingEntity> type) {
-        return SUPPLIERS.get(type);
+        AttributeSupplier supplier = net.neoforged.neoforge.common.CommonHooks.getAttributesView().get(type);
+        return supplier != null ? supplier : SUPPLIERS.get(type);
     }
 
     public static boolean hasSupplier(EntityType<?> type) {
-        return SUPPLIERS.containsKey(type);
+        return SUPPLIERS.containsKey(type) || net.neoforged.neoforge.common.CommonHooks.getAttributesView().containsKey(type);
     }
 
     public static void validate() {
