--- a/net/minecraft/world/entity/projectile/EntityThrownExpBottle.java
+++ b/net/minecraft/world/entity/projectile/EntityThrownExpBottle.java
@@ -12,6 +12,10 @@
 import net.minecraft.world.phys.MovingObjectPositionBlock;
 import net.minecraft.world.phys.Vec3D;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public class EntityThrownExpBottle extends EntityProjectileThrowable {
 
     public EntityThrownExpBottle(EntityTypes<? extends EntityThrownExpBottle> entitytypes, World world) {
@@ -42,9 +46,18 @@
         World world = this.level();
 
         if (world instanceof WorldServer worldserver) {
-            worldserver.levelEvent(2002, this.blockPosition(), -13083194);
+            // CraftBukkit - moved to after event
+            // worldserver.levelEvent(2002, this.blockPosition(), -13083194);
             int i = 3 + worldserver.random.nextInt(5) + worldserver.random.nextInt(5);
 
+            // CraftBukkit start
+            org.bukkit.event.entity.ExpBottleEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callExpBottleEvent(this, movingobjectposition, i);
+            i = event.getExperience();
+            if (event.getShowEffect()) {
+                worldserver.levelEvent(2002, this.blockPosition(), -13083194);
+            }
+            // CraftBukkit end
+
             if (movingobjectposition instanceof MovingObjectPositionBlock movingobjectpositionblock) {
                 Vec3D vec3d = movingobjectpositionblock.getDirection().getUnitVec3();
 
@@ -53,7 +66,7 @@
                 EntityExperienceOrb.awardWithDirection(worldserver, movingobjectposition.getLocation(), this.getDeltaMovement().scale(-1.0D), i);
             }
 
-            this.discard();
+            this.discard(EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
 
     }
