--- a/net/minecraft/world/entity/projectile/EntityEgg.java
+++ b/net/minecraft/world/entity/projectile/EntityEgg.java
@@ -18,6 +18,18 @@
 import net.minecraft.world.phys.MovingObjectPosition;
 import net.minecraft.world.phys.MovingObjectPositionEntity;
 
+// CraftBukkit start
+import net.minecraft.core.Holder;
+import net.minecraft.server.level.EntityPlayer;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.animal.ChickenVariant;
+import org.bukkit.entity.Ageable;
+import org.bukkit.entity.EntityType;
+import org.bukkit.entity.Player;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.player.PlayerEggThrowEvent;
+// CraftBukkit end
+
 public class EntityEgg extends EntityProjectileThrowable {
 
     private static final EntitySize ZERO_SIZED_DIMENSIONS = EntitySize.fixed(0.0F, 0.0F);
@@ -56,36 +68,68 @@
     protected void onHit(MovingObjectPosition movingobjectposition) {
         super.onHit(movingobjectposition);
         if (!this.level().isClientSide()) {
-            if (this.random.nextInt(8) == 0) {
+            // CraftBukkit start
+            boolean hatching = this.random.nextInt(8) == 0;
+            if (true) {
+            // CraftBukkit end
                 int i = 1;
 
                 if (this.random.nextInt(32) == 0) {
                     i = 4;
                 }
 
+                // CraftBukkit start
+                EntityType hatchingType = EntityType.CHICKEN;
+
+                Entity shooter = this.getOwner();
+                if (!hatching) {
+                    i = 0;
+                }
+                if (shooter instanceof EntityPlayer) {
+                    PlayerEggThrowEvent event = new PlayerEggThrowEvent((Player) shooter.getBukkitEntity(), (org.bukkit.entity.Egg) this.getBukkitEntity(), hatching, (byte) i, hatchingType);
+                    this.level().getCraftServer().getPluginManager().callEvent(event);
+
+                    i = event.getNumHatches();
+                    hatching = event.isHatching();
+                    hatchingType = event.getHatchingType();
+                    // If hatching is set to false, ensure child count is 0
+                    if (!hatching) {
+                        i = 0;
+                    }
+                }
+                // CraftBukkit end
+
                 for (int j = 0; j < i; ++j) {
-                    EntityChicken entitychicken = EntityTypes.CHICKEN.create(this.level(), EntitySpawnReason.TRIGGERED);
+                    Entity entitychicken = this.level().getWorld().makeEntity(new org.bukkit.Location(this.level().getWorld(), this.getX(), this.getY(), this.getZ(), this.getYRot(), 0.0F), hatchingType.getEntityClass()); // CraftBukkit
 
                     if (entitychicken != null) {
-                        entitychicken.setAge(-24000);
+                        // CraftBukkit start
+                        if (entitychicken.getBukkitEntity() instanceof Ageable) {
+                            ((Ageable) entitychicken.getBukkitEntity()).setBaby();
+                        }
+                        // CraftBukkit end
                         entitychicken.snapTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), 0.0F);
-                        Optional optional = Optional.ofNullable((EitherHolder) this.getItem().get(DataComponents.CHICKEN_VARIANT)).flatMap((eitherholder) -> {
+                        Optional<Holder<ChickenVariant>> optional = Optional.ofNullable((EitherHolder) this.getItem().get(DataComponents.CHICKEN_VARIANT)).flatMap((eitherholder) -> { // CraftBukkit - decompile error
                             return eitherholder.unwrap(this.registryAccess());
                         });
 
                         Objects.requireNonNull(entitychicken);
-                        optional.ifPresent(entitychicken::setVariant);
+                        // CraftBukkit start
+                        if (entitychicken instanceof EntityChicken chicken) {
+                            optional.ifPresent(chicken::setVariant);
+                        }
+                        // CraftBukkit end
                         if (!entitychicken.fudgePositionAfterSizeChange(EntityEgg.ZERO_SIZED_DIMENSIONS)) {
                             break;
                         }
 
-                        this.level().addFreshEntity(entitychicken);
+                        this.level().addFreshEntity(entitychicken, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EGG); // CraftBukkit
                     }
                 }
             }
 
             this.level().broadcastEntityEvent(this, (byte) 3);
-            this.discard();
+            this.discard(EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
 
     }
