--- a/net/minecraft/world/entity/projectile/EntityTippedArrow.java
+++ b/net/minecraft/world/entity/projectile/EntityTippedArrow.java
@@ -113,7 +113,7 @@
         float f = this.getPotionDurationScale();
 
         potioncontents.forEachEffect((mobeffect) -> {
-            entityliving.addEffect(mobeffect, entity);
+            entityliving.addEffect(mobeffect, entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
         }, f);
     }
 
