--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -118,6 +_,13 @@
         this.lookControl = new Axolotl.AxolotlLookControl(this, 20);
     }
 
+    // CraftBukkit start - SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
+    @Override
+    public int getDefaultMaxAirSupply() {
+        return AXOLOTL_TOTAL_AIR_SUPPLY;
+    }
+    // CraftBukkit end
+
     @Override
     public float getWalkTargetValue(BlockPos pos, LevelReader level) {
         return 0.0F;
@@ -270,7 +_,7 @@
 
     @Override
     public int getMaxAirSupply() {
-        return 6000;
+        return maxAirTicks; // CraftBukkit - SPIGOT-6907: re-implement LivingEntity#setMaximumAir()
     }
 
     public Axolotl.Variant getVariant() {
@@ -484,6 +_,7 @@
         if (regenEffect == null || regenEffect.endsWithin(2399)) {
             int previousDuration = regenEffect != null ? regenEffect.getDuration() : 0;
             int regenDuration = Math.min(2400, 100 + previousDuration);
+            player.addEffectCause(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.AXOLOTL); // CraftBukkit
             player.addEffect(new MobEffectInstance(MobEffects.REGENERATION, regenDuration, 0), this);
         }
 
