--- a/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/WindCharge.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/windcharge/WindCharge.java
@@ -49,14 +_,18 @@
     }
 
     @Override
-    public boolean deflect(
-        ProjectileDeflection deflection, @Nullable Entity deflectingEntity, @Nullable EntityReference<Entity> newOwner, boolean byAttack, Vec3 power
-    ) {
-        return this.noDeflectTicks > 0 ? false : super.deflect(deflection, deflectingEntity, newOwner, byAttack, power);
+    public boolean deflect(ProjectileDeflection deflection, @Nullable Entity deflectingEntity, @Nullable EntityReference<Entity> newOwner, boolean byAttack) {
+        return this.noDeflectTicks > 0 ? false : super.deflect(deflection, deflectingEntity, newOwner, byAttack);
     }
 
     @Override
     public void explode(Vec3 position) {
+        // Paper start - Fire event for WindCharge explosions
+        org.bukkit.event.entity.ExplosionPrimeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callExplosionPrimeEvent(this, RADIUS, false);
+        if (event.isCancelled()) {
+            return;
+        }
+        // Paper end - Fire event for WindCharge explosions
         this.level()
             .explode(
                 this,
@@ -65,8 +_,8 @@
                 position.x(),
                 position.y(),
                 position.z(),
-                1.2F,
-                false,
+                event.getRadius(), // Paper - Fire event for WindCharge explosions
+                event.getFire(), // Paper - Fire event for WindCharge explosions
                 Level.ExplosionInteraction.TRIGGER,
                 ParticleTypes.GUST_EMITTER_SMALL,
                 ParticleTypes.GUST_EMITTER_LARGE,
