--- a/net/minecraft/server/advancements/AdvancementVisibilityEvaluator.java
+++ b/net/minecraft/server/advancements/AdvancementVisibilityEvaluator.java
@@ -18,7 +_,7 @@
         } else if (isDone) {
             return AdvancementVisibilityEvaluator.VisibilityRule.SHOW;
         } else {
-            return display.get().hidden() ? AdvancementVisibilityEvaluator.VisibilityRule.HIDE : AdvancementVisibilityEvaluator.VisibilityRule.NO_CHANGE;
+            return display.get().isHidden() ? AdvancementVisibilityEvaluator.VisibilityRule.HIDE : AdvancementVisibilityEvaluator.VisibilityRule.NO_CHANGE;
         }
     }
 
@@ -67,6 +_,15 @@
         }
 
         evaluateVisibility(root, visibilityStack, isDone, output);
+    }
+
+    public static boolean isVisible(AdvancementNode advancement, Predicate<AdvancementNode> test) {
+        Stack<AdvancementVisibilityEvaluator.VisibilityRule> stack = new ObjectArrayList<>();
+
+        for(int i = 0; i <= 2; ++i) {
+            stack.push(AdvancementVisibilityEvaluator.VisibilityRule.NO_CHANGE);
+        }
+        return evaluateVisibility(advancement.root(), stack, test, (adv, visible) -> {});
     }
 
     @FunctionalInterface
