--- a/net/minecraft/client/color/block/BlockTintSources.java
+++ b/net/minecraft/client/color/block/BlockTintSources.java
@@ -7,7 +_,7 @@
 import net.minecraft.util.ARGB;
 import net.minecraft.world.level.GrassColor;
 import net.minecraft.world.level.block.DoublePlantBlock;
-import net.minecraft.world.level.block.RedstoneWireBlock;
+import net.minecraft.world.level.block.RedStoneWireBlock;
 import net.minecraft.world.level.block.StemBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.DoubleBlockHalf;
@@ -161,12 +_,12 @@
         return new BlockTintSource() {
             @Override
             public int color(BlockState state) {
-                return RedstoneWireBlock.getColorForPower(state.getValue(RedstoneWireBlock.POWER));
+                return RedStoneWireBlock.getColorForPower(state.getValue(RedStoneWireBlock.POWER));
             }
 
             @Override
             public Set<Property<?>> relevantProperties() {
-                return Set.of(RedstoneWireBlock.POWER);
+                return Set.of(RedStoneWireBlock.POWER);
             }
         };
     }
