--- a/net/minecraft/world/level/storage/loot/predicates/LootItemConditionTypes.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemConditionTypes.java
@@ -21,8 +_,7 @@
         Registry.register(registry, "location_check", LocationCheck.MAP_CODEC);
         Registry.register(registry, "weather_check", WeatherCheck.MAP_CODEC);
         Registry.register(registry, "time_check", TimeCheck.MAP_CODEC);
-        Registry.register(registry, "int_value_check", IntValueCheck.MAP_CODEC);
-        Registry.register(registry, "float_value_check", FloatValueCheck.MAP_CODEC);
+        Registry.register(registry, "value_check", ValueCheckCondition.MAP_CODEC);
         Registry.register(registry, "enchantment_active_check", EnchantmentActiveCheck.MAP_CODEC);
         return Registry.register(registry, "environment_attribute_check", EnvironmentAttributeCheck.MAP_CODEC);
     }
