--- a/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceCondition.java
@@ -4,13 +_,13 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.Holder;
 import net.minecraft.world.level.storage.loot.LootContext;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProvider;
-import net.minecraft.world.level.storage.loot.providers.number.floats.ContextFloatProviders;
+import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProvider;
+import net.minecraft.world.level.storage.loot.providers.number.NumberProviders;
 
-public record LootItemRandomChanceCondition(Holder<ContextFloatProvider> chance) implements LootItemCondition {
+public record LootItemRandomChanceCondition(Holder<NumberProvider> chance) implements LootItemCondition {
     public static final MapCodec<LootItemRandomChanceCondition> MAP_CODEC = RecordCodecBuilder.mapCodec(
-        i -> i.group(ContextFloatProviders.CODEC.fieldOf("chance").forGetter(LootItemRandomChanceCondition::chance))
-            .apply(i, LootItemRandomChanceCondition::new)
+        i -> i.group(NumberProviders.CODEC.fieldOf("chance").forGetter(LootItemRandomChanceCondition::chance)).apply(i, LootItemRandomChanceCondition::new)
     );
 
     @Override
@@ -24,10 +_,10 @@
     }
 
     public static LootItemCondition.Builder randomChance(float probability) {
-        return () -> new LootItemRandomChanceCondition(ContextFloatProviders.exactly(probability));
+        return () -> new LootItemRandomChanceCondition(ConstantValue.exactly(probability));
     }
 
-    public static LootItemCondition.Builder randomChance(Holder<ContextFloatProvider> probability) {
+    public static LootItemCondition.Builder randomChance(Holder<NumberProvider> probability) {
         return () -> new LootItemRandomChanceCondition(probability);
     }
 }
