--- a/net/minecraft/server/level/progress/LoggerChunkProgressListener.java
+++ b/net/minecraft/server/level/progress/LoggerChunkProgressListener.java
@@ -1,5 +_,6 @@
 package net.minecraft.server.level.progress;
 
+import com.mohistmc.youer.util.I18n;
 import com.mojang.logging.LogUtils;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
@@ -15,6 +_,7 @@
     private int count;
     private long startTime;
     private long nextTickTime = Long.MAX_VALUE;
+    private int progress = 0;
 
     private LoggerChunkProgressListener(int p_9629_) {
         this.maxCount = p_9629_;
@@ -48,7 +_,11 @@
         int i = this.getProgress();
         if (Util.getMillis() > this.nextTickTime) {
             this.nextTickTime += 500L;
-            LOGGER.info(Component.translatable("menu.preparingSpawn", Mth.clamp(i, 0, 100)).getString());
+            int s = Mth.clamp(i, 0, 100);
+            if (progress != s) {
+                progress = s;
+                LOGGER.info(Component.translatable("menu.preparingSpawn", s).getString());
+            }
         }
     }
 
@@ -58,7 +_,7 @@
 
     @Override
     public void stop() {
-        LOGGER.info("Time elapsed: {} ms", Util.getMillis() - this.startTime);
+        LOGGER.info(I18n.as("loggerchunkprogresslistener.timeelapsed", Util.getMillis() - this.startTime));
         this.nextTickTime = Long.MAX_VALUE;
     }
 
