--- a/net/minecraft/world/entity/decoration/painting/Painting.java
+++ b/net/minecraft/world/entity/decoration/painting/Painting.java
@@ -164,7 +_,27 @@
         return AABB.ofSize(position, xSize, ySize, zSize);
     }
 
-    private double offsetForPaintingSize(int size) {
+    public static AABB calculateBoundingBoxStatic(BlockPos pos, Direction direction, int width, int height) {
+        // CraftBukkit end
+        float f = 0.46875F;
+        Vec3 vec3 = Vec3.atCenterOf(pos).relative(direction, -0.46875D);
+        // CraftBukkit start
+        double d0 = offsetForPaintingSize(width);
+        double d1 = offsetForPaintingSize(height);
+        // CraftBukkit end
+        Direction direction1 = direction.getCounterClockWise();
+        Vec3 vec31 = vec3.relative(direction1, d0).relative(Direction.UP, d1);
+        Direction.Axis direction_axis = direction.getAxis();
+        // CraftBukkit start
+        double d2 = direction_axis == Direction.Axis.X ? 0.0625D : (double) width;
+        double d3 = (double) height;
+        double d4 = direction_axis == Direction.Axis.Z ? 0.0625D : (double) width;
+        // CraftBukkit end
+
+        return AABB.ofSize(vec31, d2, d3, d4);
+    }
+
+    private static double offsetForPaintingSize(int size) {
         return size % 2 == 0 ? 0.5 : 0.0;
     }
 
