--- a/net/minecraft/world/level/storage/loot/entries/TagEntry.java
+++ b/net/minecraft/world/level/storage/loot/entries/TagEntry.java
@@ -6,8 +_,8 @@
 import java.util.function.Consumer;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.storage.loot.LootContext;
@@ -16,7 +_,7 @@
 
 public class TagEntry extends ExpandableContainerBase {
     public static final MapCodec<TagEntry> MAP_CODEC = RecordCodecBuilder.mapCodec(
-        i -> i.group(RegistryCodecs.holderSet(Registries.ITEM).fieldOf("items").forGetter(e -> e.tag)).and(expandableFields(i)).apply(i, TagEntry::new)
+        i -> i.group(RegistryCodecs.homogeneousList(Registries.ITEM).fieldOf("items").forGetter(e -> e.tag)).and(expandableFields(i)).apply(i, TagEntry::new)
     );
     private final HolderSet<Item> tag;
 
