--- a/net/minecraft/server/bossevents/CustomBossEvent.java
+++ b/net/minecraft/server/bossevents/CustomBossEvent.java
@@ -18,6 +_,7 @@
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.Mth;
 import net.minecraft.world.BossEvent;
+import org.bukkit.boss.KeyedBossBar;
 
 public class CustomBossEvent extends ServerBossEvent {
     private final ResourceLocation id;
@@ -49,6 +_,10 @@
     public void removePlayer(ServerPlayer p_136281_) {
         super.removePlayer(p_136281_);
         this.players.remove(p_136281_.getUUID());
+    }
+
+    public KeyedBossBar getBukkitEntity() {
+        throw new IllegalStateException("Not implemented");
     }
 
     @Override
