--- a/net/minecraft/world/item/FireChargeItem.java
+++ b/net/minecraft/world/item/FireChargeItem.java
@@ -39,12 +39,28 @@
         if (!CampfireBlock.canLight(blockstate) && !CandleBlock.canLight(blockstate) && !CandleCakeBlock.canLight(blockstate)) {
             blockpos = blockpos.relative(useoncontext.getClickedFace());
             if (BaseFireBlock.canBePlacedAt(level, blockpos, useoncontext.getHorizontalDirection())) {
+                // CraftBukkit start - fire BlockIgniteEvent
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, blockpos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FIREBALL, useoncontext.getPlayer()).isCancelled()) {
+                    if (!useoncontext.getPlayer().getAbilities().instabuild) {
+                        useoncontext.getItemInHand().shrink(1);
+                    }
+                    return InteractionResult.PASS;
+                }
+                // CraftBukkit end
                 this.playSound(level, blockpos);
                 level.setBlockAndUpdate(blockpos, BaseFireBlock.getState(level, blockpos));
                 level.gameEvent(useoncontext.getPlayer(), (Holder) GameEvent.BLOCK_PLACE, blockpos);
                 flag = true;
             }
         } else {
+            // CraftBukkit start - fire BlockIgniteEvent
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(level, blockpos, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FIREBALL, useoncontext.getPlayer()).isCancelled()) {
+                if (!useoncontext.getPlayer().getAbilities().instabuild) {
+                    useoncontext.getItemInHand().shrink(1);
+                }
+                return InteractionResult.PASS;
+            }
+            // CraftBukkit end
             this.playSound(level, blockpos);
             level.setBlockAndUpdate(blockpos, (BlockState) blockstate.setValue(BlockStateProperties.LIT, true));
             level.gameEvent(useoncontext.getPlayer(), (Holder) GameEvent.BLOCK_CHANGE, blockpos);
