--- a/net/minecraft/world/item/ItemHanging.java
+++ b/net/minecraft/world/item/ItemHanging.java
@@ -21,6 +21,11 @@
 import net.minecraft.world.level.World;
 import net.minecraft.world.level.gameevent.GameEvent;
 
+// CraftBukkit start
+import org.bukkit.entity.Player;
+import org.bukkit.event.hanging.HangingPlaceEvent;
+// CraftBukkit end
+
 public class ItemHanging extends Item {
 
     private static final IChatBaseComponent TOOLTIP_RANDOM_VARIANT = IChatBaseComponent.translatable("painting.random").withStyle(EnumChatFormat.GRAY);
@@ -66,6 +71,19 @@
             EntityTypes.createDefaultStackConfig(world, itemstack, entityhuman).accept(entityhanging);
             if (entityhanging.survives()) {
                 if (!world.isClientSide()) {
+                    // CraftBukkit start - fire HangingPlaceEvent
+                    Player who = (itemactioncontext.getPlayer() == null) ? null : (Player) itemactioncontext.getPlayer().getBukkitEntity();
+                    org.bukkit.block.Block blockClicked = world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ());
+                    org.bukkit.block.BlockFace blockFace = org.bukkit.craftbukkit.block.CraftBlock.notchToBlockFace(enumdirection);
+                    org.bukkit.inventory.EquipmentSlot hand = org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(itemactioncontext.getHand());
+
+                    HangingPlaceEvent event = new HangingPlaceEvent((org.bukkit.entity.Hanging) entityhanging.getBukkitEntity(), who, blockClicked, blockFace, hand, org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack));
+                    world.getCraftServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        return EnumInteractionResult.FAIL;
+                    }
+                    // CraftBukkit end
                     entityhanging.playPlacementSound();
                     world.gameEvent(entityhuman, (Holder) GameEvent.ENTITY_PLACE, entityhanging.position());
                     world.addFreshEntity(entityhanging);
