--- a/net/minecraft/world/item/MobBucketItem.java
+++ b/net/minecraft/world/item/MobBucketItem.java
@@ -18,6 +18,10 @@
 import net.minecraft.world.level.material.Fluid;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.event.entity.CreatureSpawnEvent;
+// CraftBukkit end
+
 public class MobBucketItem extends BucketItem {
 
     private final EntityType<? extends Mob> type;
@@ -54,7 +58,7 @@
         }
 
         if (mob != null) {
-            level.addFreshEntityWithPassengers(mob);
+            level.addFreshEntityWithPassengers(mob, CreatureSpawnEvent.SpawnReason.BUCKET); // CraftBukkit
             mob.playAmbientSound();
         }
 
