--- a/net/minecraft/world/item/SignItem.java
+++ b/net/minecraft/world/item/SignItem.java
@@ -13,6 +13,8 @@
 
 public class SignItem extends StandingAndWallBlockItem {
 
+    public static BlockPos openSign; // CraftBukkit
+
     public SignItem(Block block, Block block1, Item.Properties item_properties) {
         super(block, block1, Direction.DOWN, item_properties);
     }
@@ -35,7 +37,10 @@
                 if (block instanceof SignBlock) {
                     SignBlock signblock = (SignBlock) block;
 
-                    signblock.openTextEdit(player, signblockentity, true);
+                    // CraftBukkit start - SPIGOT-4678
+                    // signblock.openTextEdit(player, signblockentity, true);
+                    SignItem.openSign = blockpos;
+                    // CraftBukkit end
                 }
             }
         }
