--- a/net/minecraft/world/item/SpawnEggItem.java
+++ b/net/minecraft/world/item/SpawnEggItem.java
@@ -167,7 +167,7 @@
 
     @Override
     public FeatureFlagSet requiredFeatures() {
-        return (FeatureFlagSet) Optional.ofNullable((TypedEntityData) this.components().get(DataComponents.ENTITY_DATA)).map(TypedEntityData::type).map(EntityType::requiredFeatures).orElseGet(FeatureFlagSet::of);
+        return (FeatureFlagSet) Optional.ofNullable(this.components().get(DataComponents.ENTITY_DATA)).map(TypedEntityData::type).map(EntityType::requiredFeatures).orElseGet(FeatureFlagSet::of); // CraftBukkit - decompile error
     }
 
     public Optional<Mob> spawnOffspringFromSpawnEgg(Player player, Mob mob, EntityType<? extends Mob> entitytype, ServerLevel serverlevel, Vec3 vec3, ItemStack itemstack) {
@@ -191,7 +191,7 @@
                 } else {
                     mob1.snapTo(vec3.x(), vec3.y(), vec3.z(), 0.0F, 0.0F);
                     mob1.applyComponentsFromItemStack(itemstack);
-                    serverlevel.addFreshEntityWithPassengers(mob1);
+                    serverlevel.addFreshEntityWithPassengers(mob1, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPAWNER_EGG); // CraftBukkit
                     itemstack.consume(1, player);
                     return Optional.of(mob1);
                 }
