--- a/net/minecraft/client/gui/components/toasts/AdvancementToast.java
+++ b/net/minecraft/client/gui/components/toasts/AdvancementToast.java
@@ -29,7 +_,7 @@
 
     public AdvancementToast(AdvancementHolder advancement) {
         this.advancement = advancement;
-        this.iconItem = advancement.value().display().map(d -> d.icon().create()).orElse(ItemStack.EMPTY);
+        this.iconItem = advancement.value().display().map(d -> d.getIcon().create()).orElse(ItemStack.EMPTY);
     }
 
     @Override
@@ -56,7 +_,7 @@
 
     private boolean isChallengeAdvancement() {
         Optional<DisplayInfo> displayInfo = this.advancement.value().display();
-        return displayInfo.isPresent() && displayInfo.get().type().equals(AdvancementType.CHALLENGE);
+        return displayInfo.isPresent() && displayInfo.get().getType().equals(AdvancementType.CHALLENGE);
     }
 
     @Override
@@ -64,17 +_,17 @@
         DisplayInfo display = this.advancement.value().display().orElse(null);
         graphics.blitSprite(RenderPipelines.GUI_TEXTURED, BACKGROUND_SPRITE, 0, 0, this.width(), this.height());
         if (display != null) {
-            List<FormattedCharSequence> lines = font.split(display.title(), 125);
-            int titleColor = display.type() == AdvancementType.CHALLENGE ? -30465 : -256;
+            List<FormattedCharSequence> lines = font.split(display.getTitle(), 125);
+            int titleColor = display.getType() == AdvancementType.CHALLENGE ? -30465 : -256;
             if (lines.size() == 1) {
-                graphics.text(font, display.type().getDisplayName(), 30, 7, titleColor, false);
+                graphics.text(font, display.getType().getDisplayName(), 30, 7, titleColor, false);
                 graphics.text(font, lines.get(0), 30, 18, -1, false);
             } else {
                 int unlockTextTime = 1500;
                 float unlockFadeTime = 300.0F;
                 if (fullyVisibleForMs < 1500L) {
                     int alpha = Mth.floor(Mth.clamp((float)(1500L - fullyVisibleForMs) / 300.0F, 0.0F, 1.0F) * 255.0F);
-                    graphics.text(font, display.type().getDisplayName(), 30, 11, ARGB.color(alpha, titleColor), false);
+                    graphics.text(font, display.getType().getDisplayName(), 30, 11, ARGB.color(alpha, titleColor), false);
                 } else {
                     int alpha = Mth.floor(Mth.clamp((float)(fullyVisibleForMs - 1500L) / 300.0F, 0.0F, 1.0F) * 252.0F);
                     int y = this.height() / 2 - lines.size() * 9 / 2;
