--- a/net/minecraft/data/tags/BiomeTagsProvider.java
+++ b/net/minecraft/data/tags/BiomeTagsProvider.java
@@ -12,9 +_,14 @@
 import net.minecraft.world.level.biome.MultiNoiseBiomeSourceParameterList;
 
 public class BiomeTagsProvider extends TagsProvider<Biome> {
+    /** @deprecated Forge: Use the {@linkplain #BiomeTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public BiomeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.BIOME, lookupProvider);
     }
+    public BiomeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, Registries.BIOME, lookupProvider, modId);
+    }
 
     @Override
     protected void addTags(HolderLookup.Provider registries) {
@@ -44,7 +_,6 @@
             .add(Biomes.FLOWER_FOREST)
             .add(Biomes.BIRCH_FOREST)
             .add(Biomes.OLD_GROWTH_BIRCH_FOREST)
-            .add(Biomes.DAPPLED_FOREST)
             .add(Biomes.DARK_FOREST)
             .add(Biomes.PALE_GARDEN)
             .add(Biomes.GROVE);
@@ -183,24 +_,6 @@
             .add(Biomes.SULFUR_CAVES);
         this.tag(BiomeTags.HAS_STRONGHOLD).addTag(BiomeTags.IS_OVERWORLD);
         this.tag(BiomeTags.HAS_TRIAL_CHAMBERS).addAll(overworldBiomes.stream().filter(biomeKey -> biomeKey != Biomes.DEEP_DARK));
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_BAMBOO_JUNGLE).add(Biomes.BAMBOO_JUNGLE);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_BIRCH_FOREST).add(Biomes.BIRCH_FOREST);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_CHERRY_GROVE).add(Biomes.CHERRY_GROVE);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_DAPPLED_FOREST).add(Biomes.DAPPLED_FOREST);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_FLOWER_FOREST).add(Biomes.FLOWER_FOREST);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_FOREST).add(Biomes.FOREST);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_MEADOW).add(Biomes.MEADOW);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_OLD_GROWTH_BIRCH_FOREST).add(Biomes.OLD_GROWTH_BIRCH_FOREST);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_OLD_GROWTH_PINE_TAIGA).add(Biomes.OLD_GROWTH_PINE_TAIGA);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_OLD_GROWTH_SPRUCE_TAIGA).add(Biomes.OLD_GROWTH_SPRUCE_TAIGA);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_PALE_GARDEN).add(Biomes.PALE_GARDEN);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_SAVANNA).add(Biomes.SAVANNA);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_SNOWY_TAIGA).add(Biomes.SNOWY_TAIGA);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_SPARSE_JUNGLE).add(Biomes.SPARSE_JUNGLE);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_SWAMP).add(Biomes.SWAMP);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_TAIGA).add(Biomes.TAIGA);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_WINDSWEPT_FOREST).add(Biomes.WINDSWEPT_FOREST);
-        this.tag(BiomeTags.HAS_ABANDONED_CAMP_WOODED_BADLANDS).add(Biomes.WOODED_BADLANDS);
         this.tag(BiomeTags.HAS_NETHER_FORTRESS).addTag(BiomeTags.IS_NETHER);
         this.tag(BiomeTags.HAS_NETHER_FOSSIL).add(Biomes.SOUL_SAND_VALLEY);
         this.tag(BiomeTags.HAS_BASTION_REMNANT).add(Biomes.CRIMSON_FOREST).add(Biomes.NETHER_WASTES).add(Biomes.SOUL_SAND_VALLEY).add(Biomes.WARPED_FOREST);
@@ -255,8 +_,7 @@
             .add(Biomes.WINDSWEPT_FOREST)
             .add(Biomes.WINDSWEPT_GRAVELLY_HILLS)
             .add(Biomes.WINDSWEPT_HILLS)
-            .add(Biomes.STONY_PEAKS)
-            .add(Biomes.DAPPLED_FOREST);
+            .add(Biomes.STONY_PEAKS);
         this.tag(BiomeTags.SPAWNS_WARM_VARIANT_FARM_ANIMALS)
             .add(Biomes.DESERT)
             .add(Biomes.WARM_OCEAN)
