--- a/net/minecraft/data/tags/BlockItemTagsProvider.java
+++ b/net/minecraft/data/tags/BlockItemTagsProvider.java
@@ -59,14 +_,6 @@
 
         BlockItemTagsProvider.CombinedAppender addTag(BlockItemTagId id);
 
-        default BlockItemTagsProvider.CombinedAppender addTag(BlockItemTagId... ids) {
-            for (BlockItemTagId id : ids) {
-                this.addTag(id);
-            }
-
-            return this;
-        }
-
         default BlockItemTagsProvider.CombinedAppender add(BlockItemId... ids) {
             this.addAll(Arrays.stream(ids));
             return this;
