--- a/net/minecraft/data/tags/DamageTypeTagsProvider.java
+++ b/net/minecraft/data/tags/DamageTypeTagsProvider.java
@@ -9,9 +_,14 @@
 import net.minecraft.world.damagesource.DamageTypes;
 
 public class DamageTypeTagsProvider extends TagsProvider<DamageType> {
+    /** @deprecated Forge: Use the {@linkplain #DamageTypeTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public DamageTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.DAMAGE_TYPE, lookupProvider);
     }
+    public DamageTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, Registries.DAMAGE_TYPE, lookupProvider, modId);
+    }
 
     @Override
     protected void addTags(HolderLookup.Provider registries) {
@@ -86,7 +_,6 @@
         this.tag(DamageTypeTags.IS_FREEZING).add(DamageTypes.FREEZE);
         this.tag(DamageTypeTags.IS_LIGHTNING).add(DamageTypes.LIGHTNING_BOLT);
         this.tag(DamageTypeTags.NO_ANGER).add(DamageTypes.MOB_ATTACK_NO_AGGRO);
-        this.tag(DamageTypeTags.NO_WOLF_RETALIATION).add(DamageTypes.SULFUR_CUBE_HOT);
         this.tag(DamageTypeTags.NO_IMPACT).add(DamageTypes.DROWN);
         this.tag(DamageTypeTags.ALWAYS_MOST_SIGNIFICANT_FALL).add(DamageTypes.FELL_OUT_OF_WORLD);
         this.tag(DamageTypeTags.WITHER_IMMUNE_TO).add(DamageTypes.DROWN);
@@ -210,6 +_,5 @@
             )
             .addTag(DamageTypeTags.IS_PLAYER_ATTACK);
         this.tag(DamageTypeTags.IS_MACE_SMASH).add(DamageTypes.MACE_SMASH);
-        this.tag(DamageTypeTags.BYPASSES_COOLDOWN);
     }
 }
