--- a/net/minecraft/data/tags/EntityTypeTagsProvider.java
+++ b/net/minecraft/data/tags/EntityTypeTagsProvider.java
@@ -9,8 +_,13 @@
 import net.minecraft.world.entity.EntityTypeIds;
 
 public class EntityTypeTagsProvider extends TagsProvider<EntityType<?>> {
+    /** @deprecated Forge: Use the {@linkplain #EntityTypeTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public EntityTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
-        super(output, Registries.ENTITY_TYPE, lookupProvider);
+        this(output, lookupProvider, "vanilla");
+    }
+    public EntityTypeTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, Registries.ENTITY_TYPE, lookupProvider, modId);
     }
 
     @Override
@@ -203,8 +_,7 @@
                 EntityTypeIds.DARK_OAK_BOAT,
                 EntityTypeIds.PALE_OAK_BOAT,
                 EntityTypeIds.MANGROVE_BOAT,
-                EntityTypeIds.BAMBOO_RAFT,
-                EntityTypeIds.POPLAR_BOAT
+                EntityTypeIds.BAMBOO_RAFT
             );
         this.tag(EntityTypeTags.CAN_EQUIP_SADDLE)
             .add(
@@ -273,6 +_,5 @@
             .add(EntityTypeIds.HORSE, EntityTypeIds.ZOMBIE_HORSE, EntityTypeIds.MULE, EntityTypeIds.DONKEY, EntityTypeIds.CAMEL, EntityTypeIds.CAMEL_HUSK);
         this.tag(EntityTypeTags.CANNOT_BE_AGE_LOCKED).add(EntityTypeIds.ZOMBIE_HORSE, EntityTypeIds.SKELETON_HORSE, EntityTypeIds.VILLAGER);
         this.tag(EntityTypeTags.NOT_AFFECTED_BY_GEYSERS).add(EntityTypeIds.ENDER_DRAGON);
-        this.tag(EntityTypeTags.CANNOT_BE_DISMOUNTED_BY_ITEM_USAGE).add(EntityTypeIds.INTERACTION);
     }
 }
