--- a/net/minecraft/data/tags/PotionTagsProvider.java
+++ b/net/minecraft/data/tags/PotionTagsProvider.java
@@ -59,9 +_,5 @@
                 PotionIds.SLOW_FALLING,
                 PotionIds.LONG_SLOW_FALLING
             );
-        this.tag(PotionTags.DOUSES_FIRE).add(PotionIds.WATER);
-        this.tag(PotionTags.HURTS_WATER_SENSITIVE_ENTITIES).add(PotionIds.WATER);
-        this.tag(PotionTags.EXTINGUISHES_ENTITIES).add(PotionIds.WATER);
-        this.tag(PotionTags.REHYDRATES_AXOLOTLS).add(PotionIds.WATER);
     }
 }
