--- a/net/minecraft/data/tags/StructureTagsProvider.java
+++ b/net/minecraft/data/tags/StructureTagsProvider.java
@@ -9,9 +_,14 @@
 import net.minecraft.world.level.levelgen.structure.Structure;
 
 public class StructureTagsProvider extends TagsProvider<Structure> {
+    /** @deprecated Forge: Use the {@linkplain #StructureTagsProvider(PackOutput, CompletableFuture, String) mod id variant} */
+    @Deprecated
     public StructureTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider) {
         super(output, Registries.STRUCTURE, lookupProvider);
     }
+    public StructureTagsProvider(PackOutput output, CompletableFuture<HolderLookup.Provider> lookupProvider, String modId) {
+        super(output, Registries.STRUCTURE, lookupProvider, modId);
+    }
 
     @Override
     protected void addTags(HolderLookup.Provider registries) {
@@ -32,51 +_,20 @@
             .add(BuiltinStructures.RUINED_PORTAL_OCEAN)
             .add(BuiltinStructures.RUINED_PORTAL_STANDARD)
             .add(BuiltinStructures.RUINED_PORTAL_SWAMP);
-        this.tag(StructureTags.ABANDONED_CAMP)
-            .add(BuiltinStructures.ABANDONED_CAMP_BAMBOO_JUNGLE)
-            .add(BuiltinStructures.ABANDONED_CAMP_BIRCH_FOREST)
-            .add(BuiltinStructures.ABANDONED_CAMP_CHERRY_GROVE)
-            .add(BuiltinStructures.ABANDONED_CAMP_DAPPLED_FOREST)
-            .add(BuiltinStructures.ABANDONED_CAMP_FLOWER_FOREST)
-            .add(BuiltinStructures.ABANDONED_CAMP_FOREST)
-            .add(BuiltinStructures.ABANDONED_CAMP_MEADOW)
-            .add(BuiltinStructures.ABANDONED_CAMP_OLD_GROWTH_BIRCH_FOREST)
-            .add(BuiltinStructures.ABANDONED_CAMP_OLD_GROWTH_PINE_TAIGA)
-            .add(BuiltinStructures.ABANDONED_CAMP_OLD_GROWTH_SPRUCE_TAIGA)
-            .add(BuiltinStructures.ABANDONED_CAMP_PALE_GARDEN)
-            .add(BuiltinStructures.ABANDONED_CAMP_SAVANNA)
-            .add(BuiltinStructures.ABANDONED_CAMP_SNOWY_TAIGA)
-            .add(BuiltinStructures.ABANDONED_CAMP_SPARSE_JUNGLE)
-            .add(BuiltinStructures.ABANDONED_CAMP_SWAMP)
-            .add(BuiltinStructures.ABANDONED_CAMP_TAIGA)
-            .add(BuiltinStructures.ABANDONED_CAMP_WINDSWEPT_FOREST)
-            .add(BuiltinStructures.ABANDONED_CAMP_WOODED_BADLANDS);
         this.tag(StructureTags.CATS_SPAWN_IN).add(BuiltinStructures.SWAMP_HUT);
         this.tag(StructureTags.CATS_SPAWN_AS_BLACK).add(BuiltinStructures.SWAMP_HUT);
         this.tag(StructureTags.EYE_OF_ENDER_LOCATED).add(BuiltinStructures.STRONGHOLD);
         this.tag(StructureTags.DOLPHIN_LOCATED).addTag(StructureTags.OCEAN_RUIN).addTag(StructureTags.SHIPWRECK);
-        this.tag(StructureTags.ON_WOODLAND_MANSION_MAPS).add(BuiltinStructures.WOODLAND_MANSION);
-        this.tag(StructureTags.ON_OCEAN_MONUMENT_MAPS).add(BuiltinStructures.OCEAN_MONUMENT);
+        this.tag(StructureTags.ON_WOODLAND_EXPLORER_MAPS).add(BuiltinStructures.WOODLAND_MANSION);
+        this.tag(StructureTags.ON_OCEAN_EXPLORER_MAPS).add(BuiltinStructures.OCEAN_MONUMENT);
         this.tag(StructureTags.ON_TREASURE_MAPS).add(BuiltinStructures.BURIED_TREASURE);
-        this.tag(StructureTags.ON_BURIED_TRIAL_CHAMBERS_MAPS).add(BuiltinStructures.TRIAL_CHAMBERS);
+        this.tag(StructureTags.ON_TRIAL_CHAMBERS_MAPS).add(BuiltinStructures.TRIAL_CHAMBERS);
         this.tag(StructureTags.ON_SAVANNA_VILLAGE_MAPS).add(BuiltinStructures.VILLAGE_SAVANNA);
         this.tag(StructureTags.ON_DESERT_VILLAGE_MAPS).add(BuiltinStructures.VILLAGE_DESERT);
         this.tag(StructureTags.ON_PLAINS_VILLAGE_MAPS).add(BuiltinStructures.VILLAGE_PLAINS);
         this.tag(StructureTags.ON_TAIGA_VILLAGE_MAPS).add(BuiltinStructures.VILLAGE_TAIGA);
         this.tag(StructureTags.ON_SNOWY_VILLAGE_MAPS).add(BuiltinStructures.VILLAGE_SNOWY);
-        this.tag(StructureTags.ON_SWAMP_HUT_MAPS).add(BuiltinStructures.SWAMP_HUT);
-        this.tag(StructureTags.ON_JUNGLE_PYRAMID_MAPS).add(BuiltinStructures.JUNGLE_TEMPLE);
-        this.tag(StructureTags.ON_ABANDONED_CAMP_BAMBOO_JUNGLE_MAPS).add(BuiltinStructures.ABANDONED_CAMP_BAMBOO_JUNGLE);
-        this.tag(StructureTags.ON_ABANDONED_CAMP_CHERRY_GROVE_MAPS).add(BuiltinStructures.ABANDONED_CAMP_CHERRY_GROVE);
-        this.tag(StructureTags.ON_ABANDONED_CAMP_BIRCH_FOREST_MAPS).add(BuiltinStructures.ABANDONED_CAMP_BIRCH_FOREST);
-        this.tag(StructureTags.ON_ABANDONED_CAMP_DAPPLED_FOREST_MAPS).add(BuiltinStructures.ABANDONED_CAMP_DAPPLED_FOREST);
-        this.tag(StructureTags.ON_ABANDONED_CAMP_FLOWER_FOREST_MAPS).add(BuiltinStructures.ABANDONED_CAMP_FLOWER_FOREST);
-        this.tag(StructureTags.ON_ABANDONED_CAMP_PALE_GARDEN_MAPS).add(BuiltinStructures.ABANDONED_CAMP_PALE_GARDEN);
-        this.tag(StructureTags.ON_ABANDONED_CAMP_SWAMP_MAPS).add(BuiltinStructures.ABANDONED_CAMP_SWAMP);
-        this.tag(StructureTags.ON_ABANDONED_CAMP_WINDSWEPT_FOREST_MAPS).add(BuiltinStructures.ABANDONED_CAMP_WINDSWEPT_FOREST);
-        this.tag(StructureTags.ON_ANCIENT_CITY_MAPS).add(BuiltinStructures.ANCIENT_CITY);
-        this.tag(StructureTags.ON_MINESHAFT_MAPS).add(BuiltinStructures.MINESHAFT);
-        this.tag(StructureTags.ON_DESERT_PYRAMID_MAPS).add(BuiltinStructures.DESERT_PYRAMID);
-        this.tag(StructureTags.ON_OCEAN_RUIN_WARM_MAPS).add(BuiltinStructures.OCEAN_RUIN_WARM);
+        this.tag(StructureTags.ON_SWAMP_EXPLORER_MAPS).add(BuiltinStructures.SWAMP_HUT);
+        this.tag(StructureTags.ON_JUNGLE_EXPLORER_MAPS).add(BuiltinStructures.JUNGLE_TEMPLE);
     }
 }
