--- a/net/minecraft/data/tags/VanillaItemTagsProvider.java
+++ b/net/minecraft/data/tags/VanillaItemTagsProvider.java
@@ -185,27 +_,6 @@
                 ItemIds.COPPER_PICKAXE
             );
         this.tag(ItemTags.COMPASSES).add(ItemIds.COMPASS).add(ItemIds.RECOVERY_COMPASS);
-        this.tag(ItemTags.CLONABLE_MAPS)
-            .add(
-                ItemIds.FILLED_MAP,
-                ItemIds.OCEAN_MONUMENT_MAP,
-                ItemIds.WOODLAND_MANSION_MAP,
-                ItemIds.BURIED_TRIAL_CHAMBERS_MAP,
-                ItemIds.JUNGLE_PYRAMID_MAP,
-                ItemIds.SWAMP_HUT_MAP,
-                ItemIds.DESERT_VILLAGE_MAP,
-                ItemIds.PLAINS_VILLAGE_MAP,
-                ItemIds.SAVANNA_VILLAGE_MAP,
-                ItemIds.SNOWY_VILLAGE_MAP,
-                ItemIds.TAIGA_VILLAGE_MAP,
-                ItemIds.BURIED_TREASURE_MAP,
-                ItemIds.BURIED_ANCIENT_CITY_MAP,
-                ItemIds.BURIED_MINESHAFT_MAP,
-                ItemIds.DESERT_PYRAMID_MAP,
-                ItemIds.ABANDONED_CAMP_MAP,
-                ItemIds.WARM_OCEAN_RUINS_MAP
-            );
-        this.tag(ItemTags.EXTENDABLE_MAPS).add(ItemIds.FILLED_MAP);
         this.tag(ItemTags.CREEPER_IGNITERS).add(ItemIds.FLINT_AND_STEEL).add(ItemIds.FIRE_CHARGE);
         this.tag(ItemTags.SWORDS)
             .add(ItemIds.DIAMOND_SWORD)
@@ -701,8 +_,6 @@
         this.tag(ItemTags.CAT_COLLAR_DYES).addTag(ItemTags.DYES);
         this.tag(ItemTags.WOLF_COLLAR_DYES).addTag(ItemTags.DYES);
         this.tag(ItemTags.DOUSES_CAMPFIRES).addTag(ItemTags.SHOVELS);
-        this.tag(ItemTags.BREWING_POTION_INPUTS).add(ItemIds.POTION).add(ItemIds.SPLASH_POTION).add(ItemIds.LINGERING_POTION).add(ItemIds.GLASS_BOTTLE);
-        this.tag(ItemTags.FURNACE_FUEL_BOTTOM_TAKEABLE).add(ItemIds.WATER_BUCKET).add(ItemIds.BUCKET);
     }
 
     private static ColorCollection<ResourceKey<Item>> toIds(ColorCollection<BlockItemId> ids) {
