--- a/net/minecraft/data/tags/VanillaBlockTagsProvider.java
+++ b/net/minecraft/data/tags/VanillaBlockTagsProvider.java
@@ -304,9 +_,6 @@
             .addTag(BlockTags.CAVE_VINES);
         this.tag(BlockTags.CAUSES_PERIODIC_GEYSER_ERUPTIONS).add(new BlockItemId[]{BlockItemIds.MAGMA_BLOCK});
         this.tag(BlockTags.CAUSES_CONTINUOUS_GEYSER_ERUPTIONS).add(BlockIds.LAVA);
-        this.tag(BlockTags.NETHER_PORTAL_FRAME).add(new BlockItemId[]{BlockItemIds.OBSIDIAN});
-        this.tag(BlockTags.CONDUIT_EFFECT_BLOCK)
-            .add(BlockItemIds.PRISMARINE, BlockItemIds.PRISMARINE_BRICKS, BlockItemIds.SEA_LANTERN, BlockItemIds.DARK_PRISMARINE);
         this.tag(BlockTags.INFINIBURN_OVERWORLD).add(BlockItemIds.NETHERRACK, BlockItemIds.MAGMA_BLOCK);
         this.tag(BlockTags.INFINIBURN_NETHER).addTag(BlockTags.INFINIBURN_OVERWORLD);
         this.tag(BlockTags.INFINIBURN_END).addTag(BlockTags.INFINIBURN_OVERWORLD).add(new BlockItemId[]{BlockItemIds.BEDROCK});
@@ -349,7 +_,6 @@
                 BlockItemIds.CHEST,
                 BlockItemIds.SPAWNER
             );
-        this.tag(BlockTags.UNCARVABLE).add(new BlockItemId[]{BlockItemIds.BEDROCK});
         this.tag(BlockTags.CANDLE_CAKES).add(BlockIds.CANDLE_CAKE).addAll(BlockIds.DYED_CANDLE_CAKE);
         this.tag(BlockTags.CRYSTAL_SOUND_BLOCKS).add(BlockItemIds.AMETHYST_BLOCK, BlockItemIds.BUDDING_AMETHYST);
         this.tag(BlockTags.CAULDRONS)
@@ -492,8 +_,7 @@
                 BlockItemIds.SCULK,
                 BlockItemIds.SCULK_CATALYST,
                 BlockItemIds.SCULK_VEIN,
-                BlockItemIds.SCULK_SHRIEKER,
-                BlockItemIds.STRAW_BED
+                BlockItemIds.SCULK_SHRIEKER
             );
         this.tag(BlockTags.MINEABLE_WITH_PICKAXE)
             .add(
@@ -814,9 +_,7 @@
             .addAll(toIds(BlockItemIds.GLAZED_TERRACOTTA))
             .addAll(toIds(BlockItemIds.DYED_TERRACOTTA))
             .addAll(toIds(BlockItemIds.CONCRETE))
-            .addTag(BlockTags.SPELEOTHEMS)
-            .addAll(toIds(BlockItemIds.CONCRETE_STAIRS))
-            .addAll(toIds(BlockItemIds.CONCRETE_SLAB));
+            .addTag(BlockTags.SPELEOTHEMS);
         this.tag(BlockTags.MINEABLE_WITH_SHOVEL)
             .add(
                 BlockItemIds.CLAY,
@@ -1053,8 +_,7 @@
                 BlockItemIds.CRIMSON_ROOTS,
                 BlockItemIds.LEAF_LITTER,
                 BlockItemIds.SHORT_DRY_GRASS,
-                BlockItemIds.TALL_DRY_GRASS,
-                BlockItemIds.SHELF_MUSHROOM
+                BlockItemIds.TALL_DRY_GRASS
             );
         this.tag(BlockTags.REPLACEABLE_BY_MUSHROOMS)
             .addTag(BlockTags.LEAVES)
@@ -1666,21 +_,6 @@
                 BlockItemIds.STRAW_BED
             );
         this.tag(BlockTags.AIR).add(new BlockItemId[]{BlockItemIds.AIR}).add(BlockIds.VOID_AIR, BlockIds.CAVE_AIR);
-        this.tag(BlockTags.DANGEROUS_FOR_TELEPORTATION)
-            .add(BlockIds.FIRE, BlockIds.SOUL_FIRE, BlockIds.LAVA_CAULDRON)
-            .add(
-                BlockItemIds.CAMPFIRE,
-                BlockItemIds.SOUL_CAMPFIRE,
-                BlockItemIds.CACTUS,
-                BlockItemIds.MAGMA_BLOCK,
-                BlockItemIds.SWEET_BERRY_CROP,
-                BlockItemIds.WITHER_ROSE,
-                BlockItemIds.POINTED_DRIPSTONE,
-                BlockItemIds.POWDER_SNOW
-            );
-        this.tag(BlockTags.CAT_DOES_NOT_TELEPORT_TO).addTag(BlockTags.DANGEROUS_FOR_TELEPORTATION);
-        this.tag(BlockTags.ENDERMAN_DOES_NOT_TELEPORT_TO).addTag(BlockTags.DANGEROUS_FOR_TELEPORTATION).add(new BlockItemId[]{BlockItemIds.BEDROCK});
-        this.tag(BlockTags.SHULKER_DOES_NOT_TELEPORT_TO).add(new BlockItemId[]{BlockItemIds.BEDROCK});
     }
 
     private static ColorCollection<ResourceKey<Block>> toIds(ColorCollection<BlockItemId> ids) {
