--- a/net/minecraft/data/tags/VanillaBlockItemTagsProvider.java
+++ b/net/minecraft/data/tags/VanillaBlockItemTagsProvider.java
@@ -292,7 +_,7 @@
                 BlockItemIds.POLISHED_SULFUR_SLAB,
                 BlockItemIds.SULFUR_BRICK_SLAB
             )
-            .addTag(BlockItemTags.WOOL_SLABS, BlockItemTags.CONCRETE_SLABS);
+            .addTag(BlockItemTags.WOOL_SLABS);
         this.tag(BlockItemTags.WALLS)
             .add(
                 BlockItemIds.COBBLESTONE_WALL,
@@ -379,7 +_,7 @@
                 BlockItemIds.POLISHED_SULFUR_STAIRS,
                 BlockItemIds.SULFUR_BRICK_STAIRS
             )
-            .addTag(BlockItemTags.WOOL_STAIRS, BlockItemTags.CONCRETE_STAIRS);
+            .addTag(BlockItemTags.WOOL_STAIRS);
         this.tag(BlockItemTags.ANVIL).add(BlockItemIds.ANVIL, BlockItemIds.CHIPPED_ANVIL, BlockItemIds.DAMAGED_ANVIL);
         this.tag(BlockItemTags.RAILS).add(BlockItemIds.RAIL, BlockItemIds.POWERED_RAIL, BlockItemIds.DETECTOR_RAIL, BlockItemIds.ACTIVATOR_RAIL);
         this.tag(BlockItemTags.LEAVES)
@@ -452,8 +_,6 @@
         this.tag(BlockItemTags.TERRACOTTA).add(BlockItemIds.TERRACOTTA).addAll(BlockItemIds.DYED_TERRACOTTA.asList());
         this.tag(BlockItemTags.GLAZED_TERRACOTTA).addAll(BlockItemIds.GLAZED_TERRACOTTA.asList());
         this.tag(BlockItemTags.CONCRETE).addAll(BlockItemIds.CONCRETE.asList());
-        this.tag(BlockItemTags.CONCRETE_STAIRS).addAll(BlockItemIds.CONCRETE_STAIRS.asList());
-        this.tag(BlockItemTags.CONCRETE_SLABS).addAll(BlockItemIds.CONCRETE_SLAB.asList());
         this.tag(BlockItemTags.CONCRETE_POWDERS).addAll(BlockItemIds.CONCRETE_POWDER.asList());
         this.tag(BlockItemTags.COMPLETES_FIND_TREE_TUTORIAL).addTag(BlockItemTags.LOGS).addTag(BlockItemTags.LEAVES).addTag(BlockItemTags.WART_BLOCKS);
         this.tag(BlockItemTags.SHULKER_BOXES).add(BlockItemIds.SHULKER_BOX).addAll(BlockItemIds.DYED_SHULKER_BOX.asList());
