--- a/net/minecraft/data/tags/VillagerTradesTagsProvider.java
+++ b/net/minecraft/data/tags/VillagerTradesTagsProvider.java
@@ -140,8 +_,8 @@
         this.tag(VillagerTradeTags.CARTOGRAPHER_LEVEL_3)
             .add(
                 VillagerTrades.CARTOGRAPHER_3_COMPASS_EMERALD,
-                VillagerTrades.CARTOGRAPHER_3_EMERALD_AND_COMPASS_OCEAN_MONUMENT_MAP,
-                VillagerTrades.CARTOGRAPHER_3_EMERALD_AND_COMPASS_BURIED_TRIAL_CHAMBERS_MAP
+                VillagerTrades.CARTOGRAPHER_3_EMERALD_AND_COMPASS_OCEAN_EXPLORER_MAP,
+                VillagerTrades.CARTOGRAPHER_3_EMERALD_AND_COMPASS_TRIAL_CHAMBER_MAP
             );
         this.tag(VillagerTradeTags.CARTOGRAPHER_LEVEL_4)
             .add(
