--- a/net/minecraft/util/datafix/schemas/V1460.java
+++ b/net/minecraft/util/datafix/schemas/V1460.java
@@ -272,9 +_,7 @@
                 "blocks",
                 DSL.list(DSL.optionalFields("nbt", References.BLOCK_ENTITY.in(schema))),
                 "palette",
-                DSL.list(References.BLOCK_STATE.in(schema)),
-                "palettes",
-                DSL.list(DSL.list(References.BLOCK_STATE.in(schema)))
+                DSL.list(References.BLOCK_STATE.in(schema))
             )
         );
         schema.registerType(false, References.BLOCK_NAME, () -> DSL.constType(namespacedString()));
