--- a/net/minecraft/core/cauldron/CauldronInteraction.java
+++ b/net/minecraft/core/cauldron/CauldronInteraction.java
@@ -23,11 +_,11 @@
         private final Map<TagKey<Item>, CauldronInteraction> tags = new HashMap<>();
         private final Map<Item, CauldronInteraction> items = new HashMap<>();
 
-        public void put(Item item, CauldronInteraction interaction) {
+        void put(Item item, CauldronInteraction interaction) {
             this.items.put(item, interaction);
         }
 
-        public void put(TagKey<Item> tag, CauldronInteraction interaction) {
+        void put(TagKey<Item> tag, CauldronInteraction interaction) {
             this.tags.put(tag, interaction);
         }
 
