--- a/net/minecraft/server/packs/resources/Resource.java
+++ b/net/minecraft/server/packs/resources/Resource.java
@@ -4,7 +_,6 @@
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.io.Reader;
 import java.nio.charset.StandardCharsets;
 import java.util.Optional;
 import net.minecraft.server.packs.PackResources;
@@ -48,12 +_,6 @@
 
     public BufferedReader openAsReader() throws IOException {
         return new BufferedReader(new InputStreamReader(this.open(), StandardCharsets.UTF_8));
-    }
-
-    public String readAllAsString() throws IOException {
-        try (Reader reader = this.openAsReader()) {
-            return reader.readAllAsString();
-        }
     }
 
     public ResourceMetadata metadata() throws IOException {
