--- a/net/minecraft/server/packs/metadata/pack/PackFormat.java
+++ b/net/minecraft/server/packs/metadata/pack/PackFormat.java
@@ -205,7 +_,7 @@
                         () -> context
                             + " declares support for format "
                             + majorMin
-                            + ", but game versions supporting formats 15 to "
+                            + ", but game versions supporting formats 17 to "
                             + lastPreMinorVersion
                             + " require a "
                             + oldFieldName
@@ -254,7 +_,7 @@
                             () -> context
                                 + " declares support for formats up to "
                                 + lastPreMinorVersion
-                                + ", but game versions supporting formats 15 to "
+                                + ", but game versions supporting formats 17 to "
                                 + lastPreMinorVersion
                                 + " require a pack_format field. Add \"pack_format\": "
                                 + majorMin
@@ -295,7 +_,7 @@
                         () -> context
                             + " declares support for formats up to "
                             + lastPreMinorVersion
-                            + ", but game versions supporting formats 15 to "
+                            + ", but game versions supporting formats 17 to "
                             + lastPreMinorVersion
                             + " require a pack_format field. Add \"pack_format\": "
                             + min
