--- a/net/minecraft/util/ARGB.java
+++ b/net/minecraft/util/ARGB.java
@@ -2,9 +_,7 @@
 
 import net.minecraft.world.phys.Vec3;
 import org.joml.Vector3f;
-import org.joml.Vector3fc;
 import org.joml.Vector4f;
-import org.joml.Vector4fc;
 
 public class ARGB {
     private static final int LINEAR_CHANNEL_DEPTH = 1024;
@@ -89,46 +_,14 @@
         }
     }
 
-    public static Vector3fc multiply(Vector3fc lhs, Vector3fc rhs) {
-        if (lhs.x() == 1.0F && lhs.y() == 1.0F && lhs.z() == 1.0F) {
-            return rhs;
-        } else {
-            return rhs.x() == 1.0F && rhs.y() == 1.0F && rhs.z() == 1.0F ? lhs : lhs.mul(rhs, new Vector3f());
-        }
-    }
-
-    public static Vector4fc multiply(Vector4fc lhs, Vector4fc rhs) {
-        if (lhs.x() == 1.0F && lhs.y() == 1.0F && lhs.z() == 1.0F && lhs.w() == 1.0F) {
-            return rhs;
-        } else {
-            return rhs.x() == 1.0F && rhs.y() == 1.0F && rhs.z() == 1.0F && rhs.w() == 1.0F ? lhs : lhs.mul(rhs, new Vector4f());
-        }
-    }
-
     public static int addRgb(int lhs, int rhs) {
         return color(alpha(lhs), Math.min(red(lhs) + red(rhs), 255), Math.min(green(lhs) + green(rhs), 255), Math.min(blue(lhs) + blue(rhs), 255));
     }
 
-    public static Vector3fc addRgb(Vector3fc lhs, Vector3fc rhs) {
-        return new Vector3f(Math.min(lhs.x() + rhs.x(), 1.0F), Math.min(lhs.y() + rhs.y(), 1.0F), Math.min(lhs.z() + rhs.z(), 1.0F));
-    }
-
-    public static Vector4fc addRgb(Vector4fc lhs, Vector3fc rhs) {
-        return new Vector4f(Math.min(lhs.x() + rhs.x(), 1.0F), Math.min(lhs.y() + rhs.y(), 1.0F), Math.min(lhs.z() + rhs.z(), 1.0F), lhs.w());
-    }
-
-    public static Integer subtractRgb(Integer lhs, Integer rhs) {
+    public static int subtractRgb(int lhs, int rhs) {
         return color(alpha(lhs), Math.max(red(lhs) - red(rhs), 0), Math.max(green(lhs) - green(rhs), 0), Math.max(blue(lhs) - blue(rhs), 0));
     }
 
-    public static Vector3fc subtractRgb(Vector3fc lhs, Vector3fc rhs) {
-        return new Vector3f(Math.max(lhs.x() - rhs.x(), 0.0F), Math.max(lhs.y() - rhs.y(), 0.0F), Math.max(lhs.z() - rhs.z(), 0.0F));
-    }
-
-    public static Vector4fc subtractRgb(Vector4fc lhs, Vector3fc rhs) {
-        return new Vector4f(Math.max(lhs.x() - rhs.x(), 0.0F), Math.max(lhs.y() - rhs.y(), 0.0F), Math.max(lhs.z() - rhs.z(), 0.0F), lhs.w());
-    }
-
     public static int multiplyAlpha(int color, float alphaMultiplier) {
         if (color == 0 || alphaMultiplier <= 0.0F) {
             return 0;
@@ -150,24 +_,6 @@
         );
     }
 
-    public static Vector3fc scaleRGB(Vector3fc color, float scale) {
-        return scaleRGB(color, scale, scale, scale);
-    }
-
-    public static Vector4fc scaleRGB(Vector4fc color, float scale) {
-        return scaleRGB(color, scale, scale, scale);
-    }
-
-    public static Vector3fc scaleRGB(Vector3fc color, float scaleR, float scaleG, float scaleB) {
-        return new Vector3f(Math.clamp(color.x() * scaleR, 0.0F, 1.0F), Math.clamp(color.y() * scaleG, 0.0F, 1.0F), Math.clamp(color.z() * scaleB, 0.0F, 1.0F));
-    }
-
-    public static Vector4fc scaleRGB(Vector4fc color, float scaleR, float scaleG, float scaleB) {
-        return new Vector4f(
-            Math.clamp(color.x() * scaleR, 0.0F, 1.0F), Math.clamp(color.y() * scaleG, 0.0F, 1.0F), Math.clamp(color.z() * scaleB, 0.0F, 1.0F), color.w()
-        );
-    }
-
     public static int scaleRGB(int color, int scale) {
         return color(
             alpha(color),
@@ -182,16 +_,6 @@
         return color(alpha(color), greyscale, greyscale, greyscale);
     }
 
-    public static Vector3fc greyscale(Vector3fc color) {
-        float greyscale = color.x() * 0.3F + color.y() * 0.59F + color.z() * 0.11F;
-        return new Vector3f(greyscale, greyscale, greyscale);
-    }
-
-    public static Vector4fc greyscale(Vector4fc color) {
-        float greyscale = color.x() * 0.3F + color.y() * 0.59F + color.z() * 0.11F;
-        return new Vector4f(greyscale, greyscale, greyscale, color.w());
-    }
-
     public static int alphaBlend(int destination, int source) {
         int destinationAlpha = alpha(destination);
         int sourceAlpha = alpha(source);
@@ -212,44 +_,10 @@
         );
     }
 
-    public static Vector3fc alphaBlend(Vector3fc destination, Vector4fc source) {
-        float sourceAlpha = source.w();
-        if (sourceAlpha == 0.0F) {
-            return destination;
-        }
-
-        Vector3f sourceRgb = new Vector3f().set(source);
-        return destination.lerp(sourceRgb, sourceAlpha, sourceRgb);
-    }
-
-    public static Vector4fc alphaBlend(Vector4fc destination, Vector4fc source) {
-        float destinationAlpha = destination.w();
-        float sourceAlpha = source.w();
-        if (sourceAlpha == 1.0F) {
-            return source;
-        }
-
-        if (sourceAlpha == 0.0F) {
-            return destination;
-        }
-
-        float alpha = sourceAlpha + destinationAlpha * (1.0F - sourceAlpha);
-        return new Vector4f(
-            alphaBlendChannel(alpha, sourceAlpha, destination.x(), source.x()),
-            alphaBlendChannel(alpha, sourceAlpha, destination.y(), source.y()),
-            alphaBlendChannel(alpha, sourceAlpha, destination.z(), source.z()),
-            alpha
-        );
-    }
-
     private static int alphaBlendChannel(int resultAlpha, int sourceAlpha, int destination, int source) {
         return (source * sourceAlpha + destination * (resultAlpha - sourceAlpha)) / resultAlpha;
     }
 
-    private static float alphaBlendChannel(float resultAlpha, float sourceAlpha, float destination, float source) {
-        return (source * sourceAlpha + destination * (resultAlpha - sourceAlpha)) / resultAlpha;
-    }
-
     public static int srgbLerp(float alpha, int p0, int p1) {
         int a = Mth.lerpInt(alpha, alpha(p0), alpha(p1));
         int red = Mth.lerpInt(alpha, red(p0), red(p1));
@@ -258,14 +_,6 @@
         return color(a, red, green, blue);
     }
 
-    public static Vector3fc srgbLerp(float alpha, Vector3fc p0, Vector3fc p1) {
-        return p0.lerp(p1, alpha, new Vector3f());
-    }
-
-    public static Vector4fc srgbLerp(float alpha, Vector4fc p0, Vector4fc p1) {
-        return p0.lerp(p1, alpha, new Vector4f());
-    }
-
     public static int linearLerp(float alpha, int p0, int p1) {
         return color(
             Mth.lerpInt(alpha, alpha(p0), alpha(p1)),
@@ -314,14 +_,6 @@
 
     public static int colorFromFloat(float alpha, float red, float green, float blue) {
         return color(as8BitChannel(alpha), as8BitChannel(red), as8BitChannel(green), as8BitChannel(blue));
-    }
-
-    public static int colorFromVector3f(Vector3fc color) {
-        return color(255, as8BitChannel(color.x()), as8BitChannel(color.y()), as8BitChannel(color.z()));
-    }
-
-    public static int colorFromVector4f(Vector4fc color) {
-        return color(as8BitChannel(color.w()), as8BitChannel(color.x()), as8BitChannel(color.y()), as8BitChannel(color.z()));
     }
 
     public static Vector3f vector3fFromRGB24(int color) {
