--- a/net/minecraft/util/BlockUtil.java
+++ b/net/minecraft/util/BlockUtil.java
@@ -4,7 +_,6 @@
 import com.mojang.datafixers.util.Pair;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 import it.unimi.dsi.fastutil.ints.IntStack;
-import java.util.Locale;
 import java.util.Optional;
 import java.util.function.Predicate;
 import net.minecraft.core.BlockPos;
@@ -14,8 +_,6 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class BlockUtil {
-    public static final int MAX_POSITION_DIFFERENCE_PACKING_RADIUS = 127;
-
     public static BlockUtil.FoundRectangle getLargestRectangleAround(
         BlockPos center, Direction.Axis axis1, int limit1, Direction.Axis axis2, int limit2, Predicate<BlockPos> test
     ) {
@@ -133,51 +_,6 @@
         } while (forwardState.is(bodyBlock));
 
         return forwardState.is(headBlock) ? Optional.of(forwardPos) : Optional.empty();
-    }
-
-    public static int clampedPackDifferenceInPosition(BlockPos pos, BlockPos testPos, int xRadius, int yRadius, int zRadius) {
-        int clampedXRadius = Mth.clamp(xRadius, 0, 127);
-        int clampedYRadius = Mth.clamp(yRadius, 0, 127);
-        int clampedZRadius = Mth.clamp(zRadius, 0, 127);
-        int xDiff = Mth.clamp(testPos.getX() - pos.getX(), -127, 127);
-        int yDiff = Mth.clamp(testPos.getY() - pos.getY(), -127, 127);
-        int zDiff = Mth.clamp(testPos.getZ() - pos.getZ(), -127, 127);
-        return packDifferenceInPositionInternal(clampedXRadius, clampedYRadius, clampedZRadius, xDiff, yDiff, zDiff);
-    }
-
-    public static int packDifferenceInPosition(BlockPos pos, BlockPos testPos, int xRadius, int yRadius, int zRadius) {
-        if (xRadius <= 127 && yRadius <= 127 && zRadius <= 127) {
-            int xDiff = checkDifferenceInBounds(testPos.getX(), pos.getX());
-            int yDiff = checkDifferenceInBounds(testPos.getY(), pos.getY());
-            int zDiff = checkDifferenceInBounds(testPos.getZ(), pos.getZ());
-            return packDifferenceInPositionInternal(xRadius, yRadius, zRadius, xDiff, yDiff, zDiff);
-        } else {
-            throw new IllegalArgumentException("The radius cannot be larger than 127 when packing the difference between two positions");
-        }
-    }
-
-    private static int packDifferenceInPositionInternal(int xRadius, int yRadius, int zRadius, int xDiff, int yDiff, int zDiff) {
-        return (xDiff + xRadius & 0xFF) << 16 | (yDiff + yRadius & 0xFF) << 8 | zDiff + zRadius & 0xFF;
-    }
-
-    private static int checkDifferenceInBounds(int from, int to) {
-        int difference = from - to;
-        if (difference <= 127 && difference >= -127) {
-            return difference;
-        } else {
-            throw new IllegalArgumentException(String.format(Locale.ROOT, "The difference must be within [%s,%s]", 127, -127));
-        }
-    }
-
-    public static BlockPos unpackDifferenceInPosition(BlockPos pos, int packedDiff, int xRadius, int yRadius, int zRadius) {
-        if (xRadius <= 127 && yRadius <= 127 && zRadius <= 127) {
-            int xDiff = (packedDiff >> 16 & 0xFF) - xRadius;
-            int yDiff = (packedDiff >> 8 & 0xFF) - yRadius;
-            int zDiff = (packedDiff & 0xFF) - zRadius;
-            return pos.offset(xDiff, yDiff, zDiff);
-        } else {
-            throw new IllegalArgumentException("The radius cannot be larger than 127 when unpacking the difference between two positions.");
-        }
     }
 
     public static class FoundRectangle {
