--- a/net/minecraft/util/CubicSpline.java
+++ b/net/minecraft/util/CubicSpline.java
@@ -10,19 +_,15 @@
 import it.unimi.dsi.fastutil.floats.FloatArrayList;
 import it.unimi.dsi.fastutil.floats.FloatList;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.List;
 import java.util.Locale;
-import java.util.Objects;
-import java.util.function.Consumer;
 import java.util.function.Function;
+import java.util.function.UnaryOperator;
 import java.util.stream.Collectors;
 import java.util.stream.IntStream;
 
 public sealed interface CubicSpline<I> permits CubicSpline.Multipoint, CubicSpline.Constant {
-    void forEachCoordinate(Consumer<I> consumer);
-
-    <R extends BoundedFloatFunction<?>> CubicSpline<R> mapCoordinates(Function<I, R> mapper);
+    CubicSpline<I> mapCoordinates(UnaryOperator<I> mapper);
 
     Interval range();
 
@@ -145,12 +_,8 @@
         }
 
         @Override
-        public void forEachCoordinate(Consumer<I> consumer) {
-        }
-
-        @Override
-        public <R extends BoundedFloatFunction<?>> CubicSpline<R> mapCoordinates(Function<I, R> mapper) {
-            return (CubicSpline<R>)this;
+        public CubicSpline<I> mapCoordinates(UnaryOperator<I> mapper) {
+            return this;
         }
     }
 
@@ -173,16 +_,16 @@
 
             if (inputRange.min() < this.locations[0]) {
                 Interval firstRange = this.values.getFirst().range();
-                float edge1 = linearExtend(inputRange.min(), this.locations, firstRange.min(), this.derivatives, 0);
-                float edge2 = linearExtend(inputRange.min(), this.locations, firstRange.max(), this.derivatives, 0);
+                float edge1 = linearExtend((float)inputRange.min(), this.locations, (float)firstRange.min(), this.derivatives, 0);
+                float edge2 = linearExtend((float)inputRange.min(), this.locations, (float)firstRange.max(), this.derivatives, 0);
                 minValue = Math.min(minValue, Math.min(edge1, edge2));
                 maxValue = Math.max(maxValue, Math.max(edge1, edge2));
             }
 
             if (inputRange.max() > this.locations[lastIndex]) {
                 Interval lastRange = this.values.get(lastIndex).range();
-                float edge1 = linearExtend(inputRange.max(), this.locations, lastRange.min(), this.derivatives, lastIndex);
-                float edge2 = linearExtend(inputRange.max(), this.locations, lastRange.max(), this.derivatives, lastIndex);
+                float edge1 = linearExtend((float)inputRange.max(), this.locations, (float)lastRange.min(), this.derivatives, lastIndex);
+                float edge2 = linearExtend((float)inputRange.max(), this.locations, (float)lastRange.max(), this.derivatives, lastIndex);
                 minValue = Math.min(minValue, Math.min(edge1, edge2));
                 maxValue = Math.max(maxValue, Math.max(edge1, edge2));
             }
@@ -190,8 +_,8 @@
             List<Interval> valueRanges = List.copyOf(Lists.transform(this.values, CubicSpline::range));
 
             for (Interval range : valueRanges) {
-                minValue = Math.min(minValue, range.min());
-                maxValue = Math.max(maxValue, range.max());
+                minValue = Math.min(minValue, (float)range.min());
+                maxValue = Math.max(maxValue, (float)range.max());
             }
 
             for (int i = 0; i < lastIndex; i++) {
@@ -200,10 +_,10 @@
                 float xDiff = x2 - x1;
                 Interval range1 = valueRanges.get(i);
                 Interval range2 = valueRanges.get(i + 1);
-                float min1 = range1.min();
-                float max1 = range1.max();
-                float min2 = range2.min();
-                float max2 = range2.max();
+                float min1 = (float)range1.min();
+                float max1 = (float)range1.max();
+                float min2 = (float)range2.min();
+                float max2 = (float)range2.max();
                 float d1 = this.derivatives[i];
                 float d2 = this.derivatives[i + 1];
                 if (d1 != 0.0F || d2 != 0.0F) {
@@ -300,34 +_,13 @@
         }
 
         @Override
-        public void forEachCoordinate(Consumer<I> consumer) {
-            consumer.accept(this.coordinate);
-
-            for (CubicSpline<I> spline : this.values) {
-                spline.forEachCoordinate(consumer);
-            }
-        }
-
-        @Override
-        public <R extends BoundedFloatFunction<?>> CubicSpline<R> mapCoordinates(Function<I, R> mapper) {
-            return new CubicSpline.Multipoint<>(
-                mapper.apply(this.coordinate), this.locations, this.values.stream().map(v -> v.mapCoordinates(mapper)).toList(), this.derivatives
+        public CubicSpline<I> mapCoordinates(UnaryOperator<I> mapper) {
+            return new CubicSpline.Multipoint(
+                (I)((BoundedFloatFunction)mapper.apply(this.coordinate)),
+                this.locations,
+                this.values.stream().map(v -> v.mapCoordinates(mapper)).toList(),
+                this.derivatives
             );
-        }
-
-        @Override
-        public boolean equals(Object obj) {
-            return !(obj instanceof CubicSpline.Multipoint<?> multipoint)
-                ? false
-                : Objects.equals(this.coordinate, multipoint.coordinate)
-                    && Arrays.equals(this.locations, multipoint.locations)
-                    && Arrays.equals(this.derivatives, multipoint.derivatives)
-                    && Objects.equals(this.values, multipoint.values);
-        }
-
-        @Override
-        public int hashCode() {
-            return Objects.hash(this.coordinate, Arrays.hashCode(this.locations), this.values, Arrays.hashCode(this.derivatives));
         }
 
         public static <I extends BoundedFloatFunction<?>> Codec<CubicSpline.Multipoint<I>> codec(Codec<I> coordinateCodec, Codec<CubicSpline<I>> subSplineCodec) {
