--- a/net/minecraft/util/Interval.java
+++ b/net/minecraft/util/Interval.java
@@ -1,40 +_,40 @@
 package net.minecraft.util;
 
-import it.unimi.dsi.fastutil.floats.FloatUnaryOperator;
 import java.util.List;
+import java.util.function.DoubleUnaryOperator;
 
 public final class Interval {
-    public static final Interval NaI = new Interval(Float.NaN, Float.NaN);
-    public static final Interval INFINITE = new Interval(Float.NEGATIVE_INFINITY, Float.POSITIVE_INFINITY);
-    private static final Interval NEGATIVE_ONE_TO_ONE = new Interval(-1.0F, 1.0F);
-    private static final Interval ZERO_TO_ONE = new Interval(0.0F, 1.0F);
-    private final float min;
-    private final float max;
+    public static final Interval NaI = new Interval(Double.NaN, Double.NaN);
+    public static final Interval INFINITE = new Interval(Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
+    private static final Interval NEGATIVE_ONE_TO_ONE = new Interval(-1.0, 1.0);
+    private static final Interval ZERO_TO_ONE = new Interval(0.0, 1.0);
+    private final double min;
+    private final double max;
 
-    private Interval(float min, float max) {
+    private Interval(double min, double max) {
         this.min = min;
         this.max = max;
     }
 
-    public static Interval of(float min, float max) {
+    public static Interval of(double min, double max) {
         if (max < min) {
             throw new IllegalArgumentException("max (" + max + ") < min (" + min + ")");
         }
 
-        if (Float.isNaN(min) || Float.isNaN(max)) {
+        if (Double.isNaN(min) || Double.isNaN(max)) {
             throw new IllegalArgumentException("Bounds cannot include NaN [" + min + "; " + max + "]: use Interval.NaI explicitly");
         }
 
-        if (min == Float.NEGATIVE_INFINITY && max == Float.POSITIVE_INFINITY) {
+        if (min == Double.NEGATIVE_INFINITY && max == Double.POSITIVE_INFINITY) {
             return INFINITE;
         }
 
-        if (max == 1.0F) {
-            if (min == 0.0F) {
+        if (max == 1.0) {
+            if (min == 0.0) {
                 return ZERO_TO_ONE;
             }
 
-            if (min == -1.0F) {
+            if (min == -1.0) {
                 return NEGATIVE_ONE_TO_ONE;
             }
         }
@@ -42,11 +_,11 @@
         return new Interval(min, max);
     }
 
-    public static Interval ofSymmetric(float range) {
+    public static Interval ofSymmetric(double range) {
         return of(-range, range);
     }
 
-    public static Interval ofExact(float value) {
+    public static Interval ofExact(double value) {
         return of(value, value);
     }
 
@@ -55,8 +_,8 @@
             throw new IllegalArgumentException("At least one interval required");
         }
 
-        float min = Float.POSITIVE_INFINITY;
-        float max = Float.NEGATIVE_INFINITY;
+        double min = Double.POSITIVE_INFINITY;
+        double max = Double.NEGATIVE_INFINITY;
 
         for (Interval interval : intervals) {
             if (!interval.isNaI()) {
@@ -72,60 +_,52 @@
         return encapsulating(List.of(intervals));
     }
 
-    public static Interval encapsulating(float first, float second) {
-        if (Float.isNaN(first) && Float.isNaN(second)) {
+    public static Interval encapsulating(double first, double second) {
+        if (Double.isNaN(first) && Double.isNaN(second)) {
             return NaI;
-        } else if (Float.isNaN(first)) {
+        } else if (Double.isNaN(first)) {
             return ofExact(second);
         } else {
-            return Float.isNaN(second) ? ofExact(first) : of(Math.min(first, second), Math.max(first, second));
-        }
-    }
-
-    private static Interval encapsulating(Interval first, float second) {
-        if (Float.isNaN(second)) {
-            return first;
-        } else {
-            return first.isNaI() ? ofExact(second) : of(Math.min(first.min(), second), Math.max(first.max(), second));
+            return Double.isNaN(second) ? ofExact(first) : of(Math.min(first, second), Math.max(first, second));
         }
     }
 
     public static Interval add(Interval left, Interval right) {
-        float min = left.min + right.min;
-        float max = left.max + right.max;
-        return !Float.isNaN(min) && !Float.isNaN(max) ? of(min, max) : NaI;
+        double min = left.min + right.min;
+        double max = left.max + right.max;
+        return !Double.isNaN(min) && !Double.isNaN(max) ? of(min, max) : NaI;
     }
 
     public static Interval sub(Interval left, Interval right) {
-        float min = left.min - right.max;
-        float max = left.max - right.min;
-        return !Float.isNaN(min) && !Float.isNaN(max) ? of(min, max) : NaI;
+        double min = left.min - right.max;
+        double max = left.max - right.min;
+        return !Double.isNaN(min) && !Double.isNaN(max) ? of(min, max) : NaI;
     }
 
     public static Interval mul(Interval left, Interval right) {
         if (!left.isNaI() && !right.isNaI()) {
-            float minMin = mulBound(left.min, right.min);
-            float minMax = mulBound(left.min, right.max);
-            float maxMin = mulBound(left.max, right.min);
-            float maxMax = mulBound(left.max, right.max);
+            double minMin = mulBound(left.min, right.min);
+            double minMax = mulBound(left.min, right.max);
+            double maxMin = mulBound(left.max, right.min);
+            double maxMax = mulBound(left.max, right.max);
             return of(Math.min(Math.min(minMin, minMax), Math.min(maxMin, maxMax)), Math.max(Math.max(minMin, minMax), Math.max(maxMin, maxMax)));
         } else {
             return NaI;
         }
     }
 
-    private static float mulBound(float left, float right) {
-        return left != 0.0F && right != 0.0F ? left * right : 0.0F;
+    private static double mulBound(double left, double right) {
+        return left != 0.0 && right != 0.0 ? left * right : 0.0;
     }
 
     public static Interval reciprocal(Interval input) {
-        if (!input.isNaI() && (input.min != 0.0F || input.max != 0.0F)) {
-            if (!input.contains(0.0F)) {
-                return of(1.0F / input.max, 1.0F / input.min);
-            } else if (input.max == 0.0F) {
-                return of(Float.NEGATIVE_INFINITY, 1.0F / input.min);
+        if (!input.isNaI() && (input.min != 0.0 || input.max != 0.0)) {
+            if (!input.contains(0.0)) {
+                return of(1.0 / input.max, 1.0 / input.min);
+            } else if (input.max == 0.0) {
+                return of(Double.NEGATIVE_INFINITY, 1.0 / input.min);
             } else {
-                return input.min == 0.0F ? of(1.0F / input.max, Float.POSITIVE_INFINITY) : INFINITE;
+                return input.min == 0.0 ? of(1.0 / input.max, Double.POSITIVE_INFINITY) : INFINITE;
             }
         } else {
             return NaI;
@@ -144,7 +_,7 @@
         return !left.isNaI() && !right.isNaI() ? of(Math.max(left.min, right.min), Math.max(left.max, right.max)) : NaI;
     }
 
-    public static Interval clamp(Interval input, float min, float max) {
+    public static Interval clamp(Interval input, double min, double max) {
         if (min > max) {
             throw new IllegalArgumentException("min (" + min + ") > max (" + max + ")");
         } else if (input.isNaI()) {
@@ -161,8 +_,8 @@
             return NaI;
         }
 
-        float max = Math.max(Math.abs(input.min), Math.abs(input.max));
-        return input.contains(0.0F) ? of(0.0F, max) : of(Math.min(Math.abs(input.min), Math.abs(input.max)), max);
+        double max = Math.max(Math.abs(input.min), Math.abs(input.max));
+        return input.contains(0.0) ? of(0.0, max) : of(Math.min(Math.abs(input.min), Math.abs(input.max)), max);
     }
 
     public static Interval square(Interval input) {
@@ -170,119 +_,17 @@
             return NaI;
         }
 
-        float max = Math.max(Mth.square(input.min), Mth.square(input.max));
-        return input.contains(0.0F) ? of(0.0F, max) : of(Math.min(Mth.square(input.min), Mth.square(input.max)), max);
-    }
-
-    public static Interval pow(Interval base, Interval exponent) {
-        if (!base.isNaI() && !exponent.isNaI()) {
-            if (base.min() == base.max()) {
-                return pow(base.min(), exponent);
-            }
-
-            Interval result = encapsulating(pow(base.min(), exponent), pow(base.max(), exponent));
-            if (base.contains(0.0F)) {
-                if (base.max() > 0.0F) {
-                    result = encapsulating(result, pow(0.0F, exponent));
-                }
-
-                if (base.min() < 0.0F) {
-                    result = encapsulating(result, pow(-0.0F, exponent));
-                }
-            }
-
-            return result;
-        } else {
-            return NaI;
-        }
-    }
-
-    private static Interval pow(float base, Interval exponent) {
-        if (!Float.isNaN(base) && !exponent.isNaI()) {
-            if (exponent.min() == exponent.max()) {
-                float value = (float)Math.pow(base, exponent.min());
-                return Float.isNaN(value) ? NaI : ofExact(value);
-            } else if (base == 0.0F) {
-                return mul(powZeroBase(exponent), ofExact(Math.copySign(1.0F, base)));
-            } else if (base == 1.0F) {
-                return ofExact(1.0F);
-            } else {
-                return base > 0.0F ? powPositiveBase(base, exponent) : powNegativeBase(base, exponent);
-            }
-        } else {
-            return NaI;
-        }
-    }
-
-    private static Interval powPositiveBase(float base, Interval exponent) {
-        return Float.isFinite(exponent.min()) && Float.isFinite(exponent.max())
-            ? encapsulating((float)Math.pow(base, exponent.min()), (float)Math.pow(base, exponent.max()))
-            : powInfiniteExponent(base, exponent);
-    }
-
-    private static Interval powZeroBase(Interval exponent) {
-        if (exponent.contains(0.0F)) {
-            if (exponent.max() == 0.0F) {
-                return of(1.0F, Float.POSITIVE_INFINITY);
-            } else {
-                return exponent.min() == 0.0F ? ZERO_TO_ONE : of(0.0F, Float.POSITIVE_INFINITY);
-            }
-        } else {
-            return exponent.max() < 0.0F ? ofExact(Float.POSITIVE_INFINITY) : ofExact(0.0F);
-        }
-    }
-
-    private static Interval powInfiniteExponent(float base, Interval exponent) {
-        if (Float.isInfinite(exponent.min()) && Float.isInfinite(exponent.max())) {
-            return of(0.0F, Float.POSITIVE_INFINITY);
-        } else if (Float.isInfinite(exponent.min())) {
-            return base < 1.0F ? of((float)Math.pow(base, exponent.max()), Float.POSITIVE_INFINITY) : of(0.0F, (float)Math.pow(base, exponent.max()));
-        } else {
-            return base < 1.0F ? of(0.0F, (float)Math.pow(base, exponent.min())) : of((float)Math.pow(base, exponent.min()), Float.POSITIVE_INFINITY);
-        }
-    }
-
-    private static Interval powNegativeBase(float base, Interval exponent) {
-        float exponentMinInt = (float)Math.ceil(exponent.min());
-        float exponentMaxInt = (float)Math.floor(exponent.max());
-        if (exponentMaxInt < exponentMinInt) {
-            return NaI;
-        }
-
-        float baseToMinInt = (float)Math.pow(base, exponentMinInt);
-        float baseToMaxInt = (float)Math.pow(base, exponentMaxInt);
-        Interval result = encapsulating(baseToMinInt, baseToMaxInt);
-        if (Float.isInfinite(exponentMinInt)) {
-            result = encapsulating(result, -baseToMinInt);
-        } else if (exponentMinInt + 1.0F < exponentMaxInt) {
-            result = encapsulating(result, (float)Math.pow(base, exponentMinInt + 1.0F));
-        }
-
-        if (Float.isInfinite(exponentMaxInt)) {
-            result = encapsulating(result, -baseToMaxInt);
-        } else if (exponentMaxInt - 1.0F > exponentMinInt) {
-            result = encapsulating(result, (float)Math.pow(base, exponentMaxInt - 1.0F));
-        }
-
-        return result;
-    }
-
-    public static Interval log(Interval input) {
-        if (input.max() < 0.0F) {
-            return NaI;
-        }
-
-        Interval clippedInput = max(input, ofExact(0.0F));
-        return mapMonotonic(clippedInput, x -> (float)Math.log(x));
-    }
-
-    public static Interval mapMonotonic(Interval input, FloatUnaryOperator monotonicOp) {
+        double max = Math.max(Mth.square(input.min), Mth.square(input.max));
+        return input.contains(0.0) ? of(0.0, max) : of(Math.min(Mth.square(input.min), Mth.square(input.max)), max);
+    }
+
+    public static Interval mapMonotonic(Interval input, DoubleUnaryOperator monotonicOp) {
         if (input.isNaI()) {
             return NaI;
         } else {
-            float mappedMin = monotonicOp.apply(input.min);
-            float mappedMax = monotonicOp.apply(input.max);
-            if (!Float.isNaN(mappedMin) && !Float.isNaN(mappedMax)) {
+            double mappedMin = monotonicOp.applyAsDouble(input.min);
+            double mappedMax = monotonicOp.applyAsDouble(input.max);
+            if (!Double.isNaN(mappedMin) && !Double.isNaN(mappedMax)) {
                 return of(Math.min(mappedMin, mappedMax), Math.max(mappedMin, mappedMax));
             } else {
                 throw new IllegalStateException("Monotonic operator should not produce NaN");
@@ -298,67 +_,47 @@
             : NaI;
     }
 
-    public static Interval lerp(Interval alpha, float first, float second) {
-        if (alpha.isNaI() || Float.isNaN(first) || Float.isNaN(second)) {
+    public static Interval lerp(Interval alpha, double first, double second) {
+        if (alpha.isNaI() || Double.isNaN(first) || Double.isNaN(second)) {
             return NaI;
         } else {
-            return Float.isFinite(first) && Float.isFinite(second) ? lerpFiniteBounds(alpha, first, second) : lerpInfiniteBounds(alpha, first, second);
+            return Double.isFinite(first) && Double.isFinite(second) ? lerpFiniteBounds(alpha, first, second) : lerpInfiniteBounds(alpha, first, second);
         }
     }
 
-    private static Interval lerpFiniteBounds(Interval alpha, float first, float second) {
+    private static Interval lerpFiniteBounds(Interval alpha, double first, double second) {
         return encapsulating(lerpFiniteBound(alpha.min, first, second), lerpFiniteBound(alpha.max, first, second));
     }
 
-    private static float lerpFiniteBound(float alpha, float first, float second) {
+    private static double lerpFiniteBound(double alpha, double first, double second) {
         return first + mulBound(alpha, second - first);
     }
 
-    private static Interval lerpInfiniteBounds(Interval alpha, float first, float second) {
+    private static Interval lerpInfiniteBounds(Interval alpha, double first, double second) {
         if (first == second) {
             return ofExact(first);
         }
 
-        float newMin = lerpInfiniteBound(alpha.min, first, second);
-        float newMax = lerpInfiniteBound(alpha.max, first, second);
-        return !Float.isNaN(newMin) && !Float.isNaN(newMax) ? encapsulating(newMin, newMax) : NaI;
+        double newMin = lerpInfiniteBound(alpha.min, first, second);
+        double newMax = lerpInfiniteBound(alpha.max, first, second);
+        return !Double.isNaN(newMin) && !Double.isNaN(newMax) ? encapsulating(newMin, newMax) : NaI;
     }
 
-    private static float lerpInfiniteBound(float alpha, float first, float second) {
-        float firstPart = mulBound(1.0F - alpha, first);
-        float secondPart = mulBound(alpha, second);
-        if (!Float.isInfinite(firstPart) || !Float.isInfinite(secondPart)) {
+    private static double lerpInfiniteBound(double alpha, double first, double second) {
+        double firstPart = mulBound(1.0 - alpha, first);
+        double secondPart = mulBound(alpha, second);
+        if (!Double.isInfinite(firstPart) || !Double.isInfinite(secondPart)) {
             return firstPart + secondPart;
-        } else if (alpha <= 0.0F) {
-            return second > first ? Float.NEGATIVE_INFINITY : Float.POSITIVE_INFINITY;
-        } else if (alpha >= 1.0F) {
-            return second > first ? Float.POSITIVE_INFINITY : Float.NEGATIVE_INFINITY;
-        } else {
-            return Float.NaN;
-        }
-    }
-
-    public static Interval sign(Interval input) {
-        if (input.isNaI()) {
-            return NaI;
-        }
-
-        if (input.min() == input.max()) {
-            return ofExact(Math.signum(input.min()));
-        }
-
-        if (input.contains(0.0F)) {
-            if (input.min() == 0.0F) {
-                return of(0.0F, 1.0F);
-            } else {
-                return input.max() == 0.0F ? of(-1.0F, 0.0F) : of(-1.0F, 1.0F);
-            }
-        } else {
-            return ofExact(input.min() > 0.0F ? 1.0F : -1.0F);
-        }
-    }
-
-    public boolean contains(float value) {
+        } else if (alpha <= 0.0) {
+            return second > first ? Double.NEGATIVE_INFINITY : Double.POSITIVE_INFINITY;
+        } else if (alpha >= 1.0) {
+            return second > first ? Double.POSITIVE_INFINITY : Double.NEGATIVE_INFINITY;
+        } else {
+            return Double.NaN;
+        }
+    }
+
+    public boolean contains(double value) {
         return value >= this.min && value <= this.max;
     }
 
@@ -370,11 +_,11 @@
         return this == NaI;
     }
 
-    public float min() {
+    public double min() {
         return this.min;
     }
 
-    public float max() {
+    public double max() {
         return this.max;
     }
 
@@ -385,8 +_,8 @@
 
     @Override
     public int hashCode() {
-        int hash = Float.hashCode(this.min);
-        return hash * 31 + Float.hashCode(this.max);
+        int hash = Double.hashCode(this.min);
+        return hash * 31 + Double.hashCode(this.max);
     }
 
     @Override
