--- a/net/minecraft/util/Mth.java
+++ b/net/minecraft/util/Mth.java
@@ -551,16 +_,6 @@
         return p0 + alpha1 * (p1 - p0);
     }
 
-    public static float lerp2(float alpha1, float alpha2, float x00, float x10, float x01, float x11) {
-        return lerp(alpha2, lerp(alpha1, x00, x10), lerp(alpha1, x01, x11));
-    }
-
-    public static float lerp3(
-        float alpha1, float alpha2, float alpha3, float x000, float x100, float x010, float x110, float x001, float x101, float x011, float x111
-    ) {
-        return lerp(alpha3, lerp2(alpha1, alpha2, x000, x100, x010, x110), lerp2(alpha1, alpha2, x001, x101, x011, x111));
-    }
-
     public static Vec3 lerp(double alpha, Vec3 p1, Vec3 p2) {
         return new Vec3(lerp(alpha, p1.x, p2.x), lerp(alpha, p1.y, p2.y), lerp(alpha, p1.z, p2.z));
     }
@@ -589,12 +_,12 @@
             );
     }
 
-    public static float smoothstep(float x) {
-        return x * x * x * (x * (x * 6.0F - 15.0F) + 10.0F);
+    public static double smoothstep(double x) {
+        return x * x * x * (x * (x * 6.0 - 15.0) + 10.0);
     }
 
-    public static float smoothstepDerivative(float x) {
-        return 30.0F * x * x * (x - 1.0F) * (x - 1.0F);
+    public static double smoothstepDerivative(double x) {
+        return 30.0 * x * x * (x - 1.0) * (x - 1.0);
     }
 
     public static int sign(double number) {
@@ -703,10 +_,6 @@
         return x * x + y * y;
     }
 
-    public static float lengthSquared(float x, float y) {
-        return x * x + y * y;
-    }
-
     public static double length(double x, double y) {
         return Math.sqrt(lengthSquared(x, y));
     }
@@ -721,10 +_,6 @@
 
     public static double length(double x, double y, double z) {
         return Math.sqrt(lengthSquared(x, y, z));
-    }
-
-    public static float length(float x, float y, float z) {
-        return sqrt(lengthSquared(x, y, z));
     }
 
     public static float lengthSquared(float x, float y, float z) {
