--- a/net/minecraft/util/PlaceholderLookupProvider.java
+++ b/net/minecraft/util/PlaceholderLookupProvider.java
@@ -4,12 +_,14 @@
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.DynamicOps;
 import com.mojang.serialization.JavaOps;
+import com.mojang.serialization.Lifecycle;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Optional;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderLookup;
+import net.minecraft.core.HolderOwner;
 import net.minecraft.core.HolderSet;
 import net.minecraft.core.Registry;
 import net.minecraft.resources.RegistryOps;
@@ -32,13 +_,33 @@
     }
 
     public <V> RegistryOps<V> createSerializationContext(DynamicOps<V> parent) {
-        return RegistryOps.create(parent, new RegistryOps.RegistryInfoLookup() {
-            @Override
-            public <T> Optional<HolderGetter<T>> lookup(ResourceKey<? extends Registry<? extends T>> registryKey) {
-                Optional<HolderGetter<T>> result = PlaceholderLookupProvider.this.context.lookup(registryKey).map(e -> (HolderGetter<T>)e);
-                return result.or(() -> Optional.of(PlaceholderLookupProvider.this.lookup.castAsLookup()));
+        return RegistryOps.create(
+            parent,
+            new RegistryOps.RegistryInfoLookup() {
+                @Override
+                public <T> Optional<RegistryOps.RegistryInfo<T>> lookup(ResourceKey<? extends Registry<? extends T>> registryKey) {
+                    return PlaceholderLookupProvider.this.context
+                        .lookup(registryKey)
+                        .map(RegistryOps.RegistryInfo::fromRegistryLookup)
+                        .or(
+                            () -> Optional.of(
+                                new RegistryOps.RegistryInfo<>(
+                                    PlaceholderLookupProvider.this.lookup.castAsOwner(),
+                                    PlaceholderLookupProvider.this.lookup.castAsLookup(),
+                                    Lifecycle.experimental()
+                                )
+                            )
+                        );
+                }
+
+                // Paper start add method to get the value for pre-filling builders in the reg mod API
+                @Override
+                public HolderLookup.Provider lookupForValueCopyViaBuilders() {
+                    return PlaceholderLookupProvider.this.context;
+                }
+                // Paper end - add method to get the value for pre-filling builders in the reg mod API
             }
-        });
+        );
     }
 
     public RegistryContextSwapper createSwapper() {
