--- a/net/minecraft/util/ProblemReporter.java
+++ b/net/minecraft/util/ProblemReporter.java
@@ -74,16 +_,6 @@
             return this.problems.isEmpty();
         }
 
-        public boolean hasFatalProblems() {
-            for (ProblemReporter.Collector.Entry entry : this.problems) {
-                if (entry.problem.isFatal()) {
-                    return true;
-                }
-            }
-
-            return false;
-        }
-
         public void forEach(BiConsumer<String, ProblemReporter.Problem> output) {
             List<ProblemReporter.PathElement> pathElements = new ArrayList<>();
             StringBuilder pathString = new StringBuilder();
@@ -230,10 +_,6 @@
 
     interface Problem {
         String description();
-
-        default boolean isFatal() {
-            return false;
-        }
     }
 
     record RootElementPathElement(ResourceKey<?> id) implements ProblemReporter.PathElement {
