--- a/net/minecraft/util/SimpleBitStorage.java
+++ b/net/minecraft/util/SimpleBitStorage.java
@@ -205,8 +_,8 @@
     private final long mask;
     private final int size;
     private final int valuesPerLong;
-    private final int divideMul;
-    private final int divideAdd;
+    private final int divideMul; private final long divideMulUnsigned; // Paper - Perf: Optimize SimpleBitStorage; referenced in b(int) with 2 Integer.toUnsignedLong calls
+    private final int divideAdd; private final long divideAddUnsigned; // Paper - Perf: Optimize SimpleBitStorage
     private final int divideShift;
 
     public SimpleBitStorage(int bits, int size, int[] values) {
@@ -249,8 +_,8 @@
         this.mask = (1L << bits) - 1L;
         this.valuesPerLong = (char)(64 / bits);
         int row = 3 * (this.valuesPerLong - 1);
-        this.divideMul = MAGIC[row + 0];
-        this.divideAdd = MAGIC[row + 1];
+        this.divideMul = MAGIC[row + 0]; this.divideMulUnsigned = Integer.toUnsignedLong(this.divideMul); // Paper - Perf: Optimize SimpleBitStorage
+        this.divideAdd = MAGIC[row + 1]; this.divideAddUnsigned = Integer.toUnsignedLong(this.divideAdd); // Paper - Perf: Optimize SimpleBitStorage
         this.divideShift = MAGIC[row + 2];
         int requiredLength = (size + this.valuesPerLong - 1) / this.valuesPerLong;
         if (data != null) {
@@ -265,15 +_,11 @@
     }
 
     private int cellIndex(int bitIndex) {
-        long mul = Integer.toUnsignedLong(this.divideMul);
-        long add = Integer.toUnsignedLong(this.divideAdd);
-        return (int)(bitIndex * mul + add >> 32 >> this.divideShift);
+        return (int)(bitIndex * this.divideMulUnsigned + this.divideAddUnsigned >> 32 >> this.divideShift); // Paper - Perf: Optimize SimpleBitStorage
     }
 
     @Override
-    public int getAndSet(int index, int value) {
-        Validate.inclusiveBetween(0L, this.size - 1, index);
-        Validate.inclusiveBetween(0L, this.mask, value);
+    public final int getAndSet(final int index, final int value) { // Paper - Perf: Optimize SimpleBitStorage
         int cellIndex = this.cellIndex(index);
         long cellValue = this.data[cellIndex];
         int bitIndex = (index - cellIndex * this.valuesPerLong) * this.bits;
@@ -283,9 +_,7 @@
     }
 
     @Override
-    public void set(int index, int value) {
-        Validate.inclusiveBetween(0L, this.size - 1, index);
-        Validate.inclusiveBetween(0L, this.mask, value);
+    public final void set(final int index, final int value) { // Paper - Perf: Optimize SimpleBitStorage
         int cellIndex = this.cellIndex(index);
         long cellValue = this.data[cellIndex];
         int bitIndex = (index - cellIndex * this.valuesPerLong) * this.bits;
@@ -293,7 +_,7 @@
     }
 
     @Override
-    public void fill(int value) {
+    public final void fill(final int value) { // Paper - Perf: Optimize SimpleBitStorage
         Validate.inclusiveBetween(0L, this.mask, value);
         long cellValue = value & this.mask;
         int offset = this.bits;
@@ -307,8 +_,8 @@
     }
 
     @Override
-    public int get(int index) {
-        Validate.inclusiveBetween(0L, this.size - 1, index);
+    public final int get(final int index) { // Paper - Perf: Optimize SimpleBitStorage
+        //Validate.inclusiveBetween(0L, this.size - 1, index); // Paper - Perf: Optimize SimpleBitStorage
         int cellIndex = this.cellIndex(index);
         long cellValue = this.data[cellIndex];
         int bitIndex = (index - cellIndex * this.valuesPerLong) * this.bits;
@@ -376,6 +_,67 @@
     public BitStorage copy() {
         return new SimpleBitStorage(this.bits, this.size, (long[])this.data.clone());
     }
+
+    // Paper start - block counting
+    @Override
+    public final it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<it.unimi.dsi.fastutil.shorts.ShortArrayList> moonrise$countEntries() {
+        final int valuesPerLong = this.valuesPerLong;
+        final int bits = this.bits;
+        final long mask = (1L << bits) - 1L;
+        final int size = this.size;
+
+        if (bits <= 6) {
+            final it.unimi.dsi.fastutil.shorts.ShortArrayList[] byId = new it.unimi.dsi.fastutil.shorts.ShortArrayList[1 << bits];
+            final it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<it.unimi.dsi.fastutil.shorts.ShortArrayList> ret = new it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<>(1 << bits);
+
+            int index = 0;
+
+            for (long value : this.data) {
+                int li = 0;
+                do {
+                    final int paletteIdx = (int)(value & mask);
+                    value >>= bits;
+                    ++li;
+
+                    final it.unimi.dsi.fastutil.shorts.ShortArrayList coords = byId[paletteIdx];
+                    if (coords != null) {
+                        coords.add((short)index++);
+                        continue;
+                    } else {
+                        final it.unimi.dsi.fastutil.shorts.ShortArrayList newCoords = new it.unimi.dsi.fastutil.shorts.ShortArrayList(64);
+                        byId[paletteIdx] = newCoords;
+                        newCoords.add((short)index++);
+                        ret.put(paletteIdx, newCoords);
+                        continue;
+                    }
+                } while (li < valuesPerLong && index < size);
+            }
+
+            return ret;
+        } else {
+            final it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<it.unimi.dsi.fastutil.shorts.ShortArrayList> ret = new it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<>(
+                    1 << 6
+            );
+
+            int index = 0;
+
+            for (long value : this.data) {
+                int li = 0;
+                do {
+                    final int paletteIdx = (int)(value & mask);
+                    value >>= bits;
+                    ++li;
+
+                    ret.computeIfAbsent(paletteIdx, (final int key) -> {
+                        return new it.unimi.dsi.fastutil.shorts.ShortArrayList(64);
+                    }).add((short)index++);
+                } while (li < valuesPerLong && index < size);
+            }
+
+            return ret;
+        }
+    }
+    // Paper end - block counting
 
     public static class InitializationException extends RuntimeException {
         private InitializationException(String message) {
