--- a/net/minecraft/util/ZeroBitStorage.java
+++ b/net/minecraft/util/ZeroBitStorage.java
@@ -8,31 +_,31 @@
     public static final long[] RAW = new long[0];
     private final int size;
 
-    public ZeroBitStorage(int size) {
+    public ZeroBitStorage(final int size) {
         this.size = size;
     }
 
     @Override
-    public int getAndSet(int index, int value) {
-        Validate.inclusiveBetween(0L, this.size - 1, index);
-        Validate.inclusiveBetween(0L, 0L, value);
+    public final int getAndSet(final int index, final int value) { // Paper - Perf: Optimize SimpleBitStorage
+        //Validate.inclusiveBetween(0L, this.size - 1, index); // Paper - Perf: Optimize SimpleBitStorage
+        //Validate.inclusiveBetween(0L, 0L, value); // Paper - Perf: Optimize SimpleBitStorage
         return 0;
     }
 
     @Override
-    public void set(int index, int value) {
-        Validate.inclusiveBetween(0L, this.size - 1, index);
-        Validate.inclusiveBetween(0L, 0L, value);
-    }
-
-    @Override
-    public void fill(int value) {
-        Validate.inclusiveBetween(0L, 0L, value);
-    }
-
-    @Override
-    public int get(int index) {
-        Validate.inclusiveBetween(0L, this.size - 1, index);
+    public final void set(final int index, final int value) { // Paper - Perf: Optimize SimpleBitStorage
+        //Validate.inclusiveBetween(0L, this.size - 1, index); // Paper - Perf: Optimize SimpleBitStorage
+        //Validate.inclusiveBetween(0L, 0L, value); // Paper - Perf: Optimize SimpleBitStorage
+    }
+
+    @Override
+    public void fill(final int value) {
+        //Validate.inclusiveBetween(0L, 0L, value); // Paper - Perf: Optimize SimpleBitStorage
+    }
+
+    @Override
+    public final int get(final int index) { // Paper - Perf: Optimize SimpleBitStorage
+        //Validate.inclusiveBetween(0L, this.size - 1, index); // Paper - Perf: Optimize SimpleBitStorage
         return 0;
     }
 
@@ -52,14 +_,14 @@
     }
 
     @Override
-    public void getAll(IntConsumer output) {
+    public void getAll(final IntConsumer output) {
         for (int i = 0; i < this.size; i++) {
             output.accept(0);
         }
     }
 
     @Override
-    public void unpack(int[] output) {
+    public void unpack(final int[] output) {
         Arrays.fill(output, 0, this.size, 0);
     }
 
