--- a/net/minecraft/world/entity/ai/village/poi/PoiManager.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
@@ -206,7 +_,7 @@
     }
 
     @Override
-    protected void setDirty(long sectionPos) {
+    public void setDirty(long sectionPos) {
         super.setDirty(sectionPos);
         this.distanceTracker.update(sectionPos, this.distanceTracker.getLevelFromSource(sectionPos), false);
     }
