--- a/net/minecraft/world/entity/ai/village/poi/PoiRecord.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiRecord.java
@@ -6,7 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.util.VisibleForDebug;
 
 public class PoiRecord {
@@ -90,7 +_,7 @@
         public static final Codec<PoiRecord.Packed> CODEC = RecordCodecBuilder.create(
             i -> i.group(
                     BlockPos.CODEC.fieldOf("pos").forGetter(PoiRecord.Packed::pos),
-                    RegistryCodecs.holder(Registries.POINT_OF_INTEREST_TYPE).fieldOf("type").forGetter(PoiRecord.Packed::poiType),
+                    RegistryFixedCodec.create(Registries.POINT_OF_INTEREST_TYPE).fieldOf("type").forGetter(PoiRecord.Packed::poiType),
                     Codec.INT.optionalFieldOf("free_tickets", 0).forGetter(PoiRecord.Packed::freeTickets)
                 )
                 .apply(i, PoiRecord.Packed::new)
