--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -63,6 +_,7 @@
 import net.minecraft.util.Mth;
 import net.minecraft.util.debugchart.LocalSampleLogger;
 import org.apache.commons.lang3.Validate;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.slf4j.Logger;
 import org.slf4j.Marker;
 import org.slf4j.MarkerFactory;
@@ -107,6 +_,30 @@
     private volatile DisconnectionDetails delayedDisconnect;
     @Nullable
     BandwidthDebugMonitor bandwidthDebugMonitor;
+    public String hostname = ""; // CraftBukkit - add field
+    @Nullable
+    private ProtocolInfo<?> inboundProtocol;
+    // Spigot Start
+    public java.util.UUID spoofedUUID;
+    public com.mojang.authlib.properties.Property[] spoofedProfile;
+    public boolean preparing = true;
+    // Spigot End
+    private volatile boolean mohist$isClosing; //mohist
+    // Paper start - NetworkClient implementation
+    public int protocolVersion;
+    public java.net.InetSocketAddress virtualHost;
+    // Paper end
+    // Paper start - add utility methods
+    public final net.minecraft.server.level.ServerPlayer getPlayer() {
+        if (this.packetListener instanceof net.minecraft.server.network.ServerGamePacketListenerImpl impl) {
+            return impl.player;
+        } else if (this.packetListener instanceof net.minecraft.server.network.ServerCommonPacketListenerImpl impl) {
+            CraftPlayer player = impl.getCraftPlayer();
+            return player == null ? null : player.getHandle();
+        }
+        return null;
+    }
+    // Paper end - add utility methods
 
     public Connection(PacketFlow p_129482_) {
         this.receiving = p_129482_;
@@ -117,14 +_,19 @@
         super.channelActive(p_129525_);
         this.channel = p_129525_.channel();
         this.address = this.channel.remoteAddress();
+        // Spigot Start
+        this.preparing = false;
+        // Spigot End
         if (this.delayedDisconnect != null) {
             this.disconnect(this.delayedDisconnect);
         }
+        net.neoforged.neoforge.network.connection.ConnectionUtils.setConnection(p_129525_, this);
     }
 
     @Override
     public void channelInactive(ChannelHandlerContext p_129527_) {
         this.disconnect(Component.translatable("disconnect.endOfStream"));
+        net.neoforged.neoforge.network.connection.ConnectionUtils.removeConnection(p_129527_);
     }
 
     @Override
@@ -134,13 +_,20 @@
         } else {
             boolean flag = !this.handlingFault;
             this.handlingFault = true;
-            if (this.channel.isOpen()) {
+            if (this.channel.isOpen() && !this.mohist$isClosing) {
                 if (p_129534_ instanceof TimeoutException) {
                     LOGGER.debug("Timeout", p_129534_);
                     this.disconnect(Component.translatable("disconnect.timeout"));
                 } else {
                     Component component = Component.translatable("disconnect.genericReason", "Internal Exception: " + p_129534_);
                     PacketListener packetlistener = this.packetListener;
+                    if (packetlistener != null) {
+                        ConnectionProtocol protocol = packetlistener.protocol();
+                        if (protocol == ConnectionProtocol.CONFIGURATION || protocol == ConnectionProtocol.PLAY) {
+                            // Neo: Always log critical network exceptions for config and play packets
+                            LOGGER.error("Exception caught in connection", p_129534_);
+                        }
+                    }
                     DisconnectionDetails disconnectiondetails;
                     if (packetlistener != null) {
                         disconnectiondetails = packetlistener.createDisconnectionInfo(component, p_129534_);
@@ -170,7 +_,7 @@
     }
 
     protected void channelRead0(ChannelHandlerContext p_129487_, Packet<?> p_129488_) {
-        if (this.channel.isOpen()) {
+        if (this.channel.isOpen() && !this.mohist$isClosing) {
             PacketListener packetlistener = this.packetListener;
             if (packetlistener == null) {
                 throw new IllegalStateException("Received a packet before the packet listener was initialized");
@@ -226,6 +_,7 @@
         if (p_320903_.flow() != this.getReceiving()) {
             throw new IllegalStateException("Invalid inbound protocol: " + p_320903_.id());
         } else {
+            this.inboundProtocol = p_320903_;
             this.packetListener = p_320940_;
             this.disconnectListener = null;
             UnconfiguredPipelineHandler.InboundConfigurationTask unconfiguredpipelinehandler$inboundconfigurationtask = UnconfiguredPipelineHandler.setupInboundProtocol(
@@ -254,7 +_,12 @@
             if (bundlerinfo != null) {
                 PacketBundleUnpacker packetbundleunpacker = new PacketBundleUnpacker(bundlerinfo);
                 unconfiguredpipelinehandler$outboundconfigurationtask = unconfiguredpipelinehandler$outboundconfigurationtask.andThen(
-                    p_319516_ -> p_319516_.pipeline().addAfter("encoder", "unbundler", packetbundleunpacker)
+                    p_319516_ -> {
+                        p_319516_.pipeline().addAfter("encoder", "unbundler", packetbundleunpacker);
+                        // Neo: our handlers must be between the encoder and the unbundler, so re-inject them
+                        // Note, this call must be inside the .andThen lambda, or it will actually run before the unbundler gets added.
+                        net.neoforged.neoforge.network.filters.NetworkFilters.injectIfNecessary(this);
+                    }
                 );
             }
 
@@ -379,7 +_,7 @@
     }
 
     private void flushQueue() {
-        if (this.channel != null && this.channel.isOpen()) {
+        if (this.channel != null && this.channel.isOpen() && !this.mohist$isClosing) {
             synchronized (this.pendingActions) {
                 Consumer<Connection> consumer;
                 while ((consumer = this.pendingActions.poll()) != null) {
@@ -427,7 +_,7 @@
         if (this.address == null) {
             return "local";
         } else {
-            return p_295875_ ? this.address.toString() : "IP hidden";
+            return p_295875_ ? net.neoforged.neoforge.network.DualStackUtils.getAddressString(this.address) : "IP hidden";
         }
     }
 
@@ -436,12 +_,16 @@
     }
 
     public void disconnect(DisconnectionDetails p_350867_) {
+        // Spigot Start
+        this.preparing = false;
+        // Spigot End
         if (this.channel == null) {
             this.delayedDisconnect = p_350867_;
         }
 
         if (this.isConnected()) {
-            this.channel.close().awaitUninterruptibly();
+            this.mohist$isClosing = true; // Mohist fix vmp mixin
+            this.channel.close(); // We can't wait as this may be called from an event loop.
             this.disconnectionDetails = p_350867_;
         }
     }
@@ -470,6 +_,7 @@
     }
 
     public static ChannelFuture connect(InetSocketAddress p_290034_, boolean p_290035_, final Connection p_290031_) {
+        net.neoforged.neoforge.network.DualStackUtils.checkIPv6(p_290034_.getAddress());
         Class<? extends SocketChannel> oclass;
         EventLoopGroup eventloopgroup;
         if (Epoll.isAvailable() && p_290035_) {
@@ -563,7 +_,7 @@
     }
 
     public boolean isConnected() {
-        return this.channel != null && this.channel.isOpen();
+        return this.channel != null && this.channel.isOpen() && !this.mohist$isClosing;
     }
 
     public boolean isConnecting() {
@@ -624,6 +_,7 @@
                     );
                     packetlistener1.onDisconnect(disconnectiondetails);
                 }
+                this.pendingActions.clear(); // Free up packet queue.
             }
         }
     }
@@ -638,5 +_,17 @@
 
     public void setBandwidthLogger(LocalSampleLogger p_323799_) {
         this.bandwidthDebugMonitor = new BandwidthDebugMonitor(p_323799_);
+    }
+
+    public Channel channel() {
+        return this.channel;
+    }
+
+    public PacketFlow getDirection() {
+        return this.receiving;
+    }
+
+    public ProtocolInfo<?> getInboundProtocol() {
+        return Objects.requireNonNull(this.inboundProtocol, "Inbound protocol not set?");
     }
 }
