--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -7,6 +_,7 @@
 import io.netty.channel.ChannelException;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandler;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelInboundHandler;
 import io.netty.channel.ChannelInitializer;
@@ -68,6 +_,11 @@
     private final Queue<Consumer<Connection>> pendingActions = Queues.newConcurrentLinkedQueue();
     public Channel channel;
     public SocketAddress address;
+    // Spigot Start
+    public java.util.UUID spoofedUUID;
+    public com.mojang.authlib.properties.Property[] spoofedProfile;
+    public boolean preparing = true;
+    // Spigot End
     private volatile @Nullable PacketListener disconnectListener;
     private volatile @Nullable PacketListener packetListener;
     private @Nullable DisconnectionDetails disconnectionDetails;
@@ -81,6 +_,50 @@
     private volatile @Nullable DisconnectionDetails delayedDisconnect;
     private @Nullable BandwidthDebugMonitor bandwidthDebugMonitor;
     private @Nullable UUID intendedProfileId;
+    @Nullable
+    private ProtocolInfo<?> inboundProtocol;
+    public String hostname = ""; // CraftBukkit - add field
+    public boolean handleConnectionDisconnectOnNextTick = false; // Paper - Force kill connection ticking
+    // Paper start - NetworkClient implementation
+    public int protocolVersion;
+    public java.net.InetSocketAddress virtualHost;
+    private static boolean enableExplicitFlush = Boolean.getBoolean("paper.explicit-flush"); // Paper - Disable explicit connection flushing
+    // Paper end
+    // Paper start - add utility methods
+    public final net.minecraft.server.level.ServerPlayer getPlayer() {
+        if (this.packetListener instanceof net.minecraft.server.network.ServerGamePacketListenerImpl impl) {
+            return impl.player;
+        } else if (this.packetListener instanceof net.minecraft.server.network.ServerCommonPacketListenerImpl impl) {
+            return null;
+        }
+        return null;
+    }
+    // Paper end - add utility methods
+    // Paper start - packet limiter
+    protected final Object PACKET_LIMIT_LOCK = new Object();
+    protected final io.papermc.paper.util.@Nullable IntervalledCounter allPacketCounts = io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.allPackets.isEnabled() ? new io.papermc.paper.util.IntervalledCounter(
+            (long)(io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.allPackets.interval() * 1.0e9)
+    ) : null;
+    protected final java.util.Map<Class<? extends net.minecraft.network.protocol.Packet<?>>, io.papermc.paper.util.IntervalledCounter> packetSpecificLimits = new java.util.HashMap<>();
+
+    private boolean stopReadingPackets;
+    private void killForPacketSpam() {
+        this.sendPacket(new ClientboundDisconnectPacket(io.papermc.paper.adventure.PaperAdventure.asVanilla(io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.kickMessage)), PacketSendListener.thenRun(() -> {
+            this.disconnect(io.papermc.paper.adventure.PaperAdventure.asVanilla(io.papermc.paper.configuration.GlobalConfiguration.get().packetLimiter.kickMessage));
+        }), true);
+        this.setReadOnly();
+        this.stopReadingPackets = true;
+    }
+    // Paper end - packet limiter
+    @Nullable public SocketAddress haProxyAddress; // Paper - Add API to get player's proxy address
+    public java.util.@Nullable Optional<net.minecraft.network.chat.Component> legacySavedLoginEventResultOverride; // Paper - playerloginevent
+    public boolean handledLegacyLoginEvent; // Paper - playerloginevent
+    public net.minecraft.server.level.@Nullable ServerPlayer savedPlayerForLegacyEvents; // Paper - playerloginevent & PlayerSpawnLocationEvent
+    public org.bukkit.event.player.PlayerResourcePackStatusEvent.@Nullable Status resourcePackStatus; // Paper
+    // Paper start - Optimize network
+    public boolean isPending = true;
+    public boolean queueImmunity;
+    // Paper end - Optimize network
 
     public Connection(PacketFlow receiving) {
         this.receiving = receiving;
@@ -91,14 +_,19 @@
         super.channelActive(ctx);
         this.channel = ctx.channel();
         this.address = this.channel.remoteAddress();
+        // Spigot Start
+        this.preparing = false;
+        // Spigot End
         if (this.delayedDisconnect != null) {
             this.disconnect(this.delayedDisconnect);
         }
+        net.neoforged.neoforge.network.connection.ConnectionUtils.setConnection(ctx, this);
     }
 
     @Override
     public void channelInactive(ChannelHandlerContext ctx) {
         this.disconnect(Component.translatable("disconnect.endOfStream"));
+        net.neoforged.neoforge.network.connection.ConnectionUtils.removeConnection(ctx);
     }
 
     @Override
@@ -115,6 +_,13 @@
                 } else {
                     Component reason = Component.translatable("disconnect.genericReason", "Internal Exception: " + cause);
                     PacketListener listener = this.packetListener;
+                    if (listener != null) {
+                        ConnectionProtocol protocol = listener.protocol();
+                        if (protocol == ConnectionProtocol.CONFIGURATION || protocol == ConnectionProtocol.PLAY) {
+                            // Neo: Always log critical network exceptions for config and play packets
+                            LOGGER.error("Exception caught in connection", cause);
+                        }
+                    }
                     DisconnectionDetails details;
                     if (listener != null) {
                         details = listener.createDisconnectionInfo(reason, cause);
@@ -201,6 +_,7 @@
             throw new IllegalStateException("Invalid inbound protocol: " + protocol.id());
         }
 
+        this.inboundProtocol = protocol;
         this.packetListener = packetListener;
         this.disconnectListener = null;
         UnconfiguredPipelineHandler.InboundConfigurationTask configMessage = UnconfiguredPipelineHandler.setupInboundProtocol(protocol);
@@ -222,7 +_,12 @@
         BundlerInfo bundlerInfo = protocol.bundlerInfo();
         if (bundlerInfo != null) {
             PacketBundleUnpacker newUnbundler = new PacketBundleUnpacker(bundlerInfo);
-            configMessage = configMessage.andThen(ctx -> ctx.pipeline().addAfter("encoder", "unbundler", newUnbundler));
+            configMessage = configMessage.andThen(ctx -> {
+                ctx.pipeline().addAfter("encoder", "unbundler", newUnbundler);
+                // Neo: our handlers must be between the encoder and the unbundler, so re-inject them
+                // Note, this call must be inside the .andThen lambda, or it will actually run before the unbundler gets added.
+                net.neoforged.neoforge.network.filters.NetworkFilters.injectIfNecessary(this);
+            });
         }
 
         boolean isLoginProtocol = protocol.id() == ConnectionProtocol.LOGIN;
@@ -381,7 +_,7 @@
         if (this.address == null) {
             return "local";
         } else {
-            return logIPs ? this.address.toString() : "IP hidden";
+            return logIPs ? net.neoforged.neoforge.network.DualStackUtils.getAddressString(this.address) : "IP hidden";
         }
     }
 
@@ -390,12 +_,15 @@
     }
 
     public void disconnect(DisconnectionDetails details) {
+        // Spigot Start
+        this.preparing = false;
+        // Spigot End
         if (this.channel == null) {
             this.delayedDisconnect = details;
         }
 
         if (this.isConnected()) {
-            this.channel.close().awaitUninterruptibly();
+            this.channel.close(); // We can't wait as this may be called from an event loop.
             this.disconnectionDetails = details;
         }
     }
@@ -424,6 +_,7 @@
     }
 
     public static ChannelFuture connect(InetSocketAddress address, EventLoopGroupHolder eventLoopGroupHolder, Connection connection) {
+        net.neoforged.neoforge.network.DualStackUtils.checkIPv6(address.getAddress());
         return new Bootstrap().group(eventLoopGroupHolder.eventLoopGroup()).handler(new ChannelInitializer<Channel>() {
             @Override
             protected void initChannel(Channel channel) {
@@ -511,10 +_,24 @@
         return connection;
     }
 
-    public void setEncryptionKey(Cipher decryptCipher, Cipher encryptCipher) {
-        this.channel.pipeline().addBefore("splitter", "decrypt", new CipherDecoder(decryptCipher));
-        this.channel.pipeline().addBefore("prepender", "encrypt", new CipherEncoder(encryptCipher));
+    // Paper start - Use Velocity cipher
+    private boolean encrypted;
+
+    public void setEncryptionKey(final javax.crypto.SecretKey key) throws net.minecraft.util.CryptException {
+        if (!this.encrypted) {
+            try {
+                final com.velocitypowered.natives.encryption.VelocityCipher decryptionCipher = com.velocitypowered.natives.util.Natives.cipher.get().forDecryption(key);
+                final com.velocitypowered.natives.encryption.VelocityCipher encryptionCipher = com.velocitypowered.natives.util.Natives.cipher.get().forEncryption(key);
+
+                this.encrypted = true;
+                this.channel.pipeline().addBefore(HandlerNames.SPLITTER, HandlerNames.DECRYPT, new CipherDecoder(decryptionCipher));
+                this.channel.pipeline().addBefore(HandlerNames.PREPENDER, HandlerNames.ENCRYPT, new CipherEncoder(encryptionCipher));
+            } catch (final java.security.GeneralSecurityException e) {
+                throw new net.minecraft.util.CryptException(e);
+            }
+        }
     }
+    // Paper end - Use Velocity cipher
 
     public boolean isConnected() {
         return this.channel != null && this.channel.isOpen();
@@ -538,6 +_,14 @@
         }
     }
 
+    // Paper start - add proper async disconnect
+    public void enableAutoRead() {
+        if (this.channel != null) {
+            this.channel.config().setAutoRead(true);
+        }
+    }
+    // Paper end - add proper async disconnect
+
     public void setupCompression(int threshold, boolean validateDecompressed) {
         if (threshold >= 0) {
             if (this.channel.pipeline().get("decompress") instanceof CompressionDecoder compressionDecoder) {
@@ -576,6 +_,7 @@
                     );
                     disconnectListener.onDisconnect(details);
                 }
+                this.pendingActions.clear(); // Free up packet queue.
             }
         }
     }
@@ -598,5 +_,17 @@
 
     public @Nullable UUID getIntendedProfileId() {
         return this.intendedProfileId;
+    }
+
+    public Channel channel() {
+        return this.channel;
+    }
+
+    public PacketFlow getDirection() {
+        return this.receiving;
+    }
+
+    public ProtocolInfo<?> getInboundProtocol() {
+        return Objects.requireNonNull(this.inboundProtocol, "Inbound protocol not set?");
     }
 }
