--- a/net/minecraft/network/FriendlyByteBuf.java
+++ b/net/minecraft/network/FriendlyByteBuf.java
@@ -526,7 +_,7 @@
 
         try {
             NbtIo.writeAnyTag(tag, new ByteBufOutputStream(output));
-        } catch (IOException var3) {
+        } catch (Exception var3) { // CraftBukkit - IOException -> Exception
             throw new EncoderException(var3);
         }
     }
@@ -1610,5 +_,10 @@
     @Override
     public boolean release(int decrement) {
         return this.source.release(decrement);
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public ByteBuf getSource() {
+        return this.source;
     }
 }
