--- a/net/minecraft/network/FriendlyByteBuf.java
+++ b/net/minecraft/network/FriendlyByteBuf.java
@@ -511,7 +_,7 @@
 
         try {
             NbtIo.writeAnyTag(p_320692_, new ByteBufOutputStream(p_320344_));
-        } catch (IOException ioexception) {
+        } catch (Exception ioexception) { // CraftBukkit - IOException -> Exception
             throw new EncoderException(ioexception);
         }
     }
@@ -1588,5 +_,10 @@
     @Override
     public boolean release(int p_130347_) {
         return this.source.release(p_130347_);
+    }
+
+    @org.jetbrains.annotations.ApiStatus.Internal
+    public ByteBuf getSource() {
+        return this.source;
     }
 }
