--- a/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
@@ -22,7 +_,7 @@
 
 public class BarrelBlockEntity extends RandomizableContainerBlockEntity {
     private NonNullList<ItemStack> items = NonNullList.withSize(27, ItemStack.EMPTY);
-    public final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
+    private final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
         @Override
         protected void onOpen(Level p_155062_, BlockPos p_155063_, BlockState p_155064_) {
             BarrelBlockEntity.this.playSound(p_155064_, SoundEvents.BARREL_OPEN);
@@ -116,11 +_,11 @@
         }
     }
 
-    public void updateBlockState(BlockState p_58607_, boolean p_58608_) {
+    void updateBlockState(BlockState p_58607_, boolean p_58608_) {
         this.level.setBlock(this.getBlockPos(), p_58607_.setValue(BarrelBlock.OPEN, Boolean.valueOf(p_58608_)), 3);
     }
 
-    public void playSound(BlockState p_58601_, SoundEvent p_58602_) {
+    void playSound(BlockState p_58601_, SoundEvent p_58602_) {
         Vec3i vec3i = p_58601_.getValue(BarrelBlock.FACING).getNormal();
         double d0 = (double)this.worldPosition.getX() + 0.5 + (double)vec3i.getX() / 2.0;
         double d1 = (double)this.worldPosition.getY() + 0.5 + (double)vec3i.getY() / 2.0;
