--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -68,17 +_,45 @@
     protected abstract Component getDefaultName();
 
     public boolean canOpen(Player p_58645_) {
-        return canUnlock(p_58645_, this.lockKey, this.getDisplayName());
-    }
-
-    public static boolean canUnlock(Player p_58630_, LockCode p_58631_, Component p_58632_) {
-        if (!p_58630_.isSpectator() && !p_58631_.unlocksWith(p_58630_.getMainHandItem())) {
-            p_58630_.displayClientMessage(Component.translatable("container.isLocked", p_58632_), true);
-            p_58630_.playNotifySound(SoundEvents.CHEST_LOCKED, SoundSource.BLOCKS, 1.0F, 1.0F);
-            return false;
-        } else {
-            return true;
-        }
+        return BaseContainerBlockEntity.canUnlock(p_58645_, this.lockKey, this.getDisplayName(), this); // Paper - Add BlockLockCheckEvent
+    }
+
+    @Deprecated @io.papermc.paper.annotation.DoNotUse // Paper - Add BlockLockCheckEvent
+    public static boolean canUnlock(Player player, LockCode lock, Component containerName) {
+        // Paper start - Add BlockLockCheckEvent
+        return canUnlock(player, lock, containerName, null);
+    }
+
+    public static boolean canUnlock(Player p_58630_, LockCode p_58631_, Component p_58632_, @Nullable BlockEntity blockEntity) {
+        if (p_58630_ instanceof net.minecraft.server.level.ServerPlayer serverPlayer && blockEntity != null && blockEntity.getLevel() != null && blockEntity.getLevel().getBlockEntity(blockEntity.getBlockPos()) == blockEntity) {
+            final org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(blockEntity.getLevel(), blockEntity.getBlockPos());
+            net.kyori.adventure.text.Component lockedMessage = net.kyori.adventure.text.Component.translatable("container.isLocked", io.papermc.paper.adventure.PaperAdventure.asAdventure(p_58632_));
+            net.kyori.adventure.sound.Sound lockedSound = net.kyori.adventure.sound.Sound.sound(org.bukkit.Sound.BLOCK_CHEST_LOCKED, net.kyori.adventure.sound.Sound.Source.BLOCK, 1.0F, 1.0F);
+            final io.papermc.paper.event.block.BlockLockCheckEvent event = new io.papermc.paper.event.block.BlockLockCheckEvent(block, serverPlayer.getBukkitEntity(), lockedMessage, lockedSound);
+            event.callEvent();
+            if (event.getResult() == org.bukkit.event.Event.Result.ALLOW) {
+                return true;
+            } else if (event.getResult() == org.bukkit.event.Event.Result.DENY || (!p_58630_.isSpectator() && !p_58631_.unlocksWith(event.isUsingCustomKeyItemStack() ? org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getKeyItem()) : p_58630_.getMainHandItem()))) {
+                if (event.getLockedMessage() != null) {
+                    event.getPlayer().sendActionBar(event.getLockedMessage());
+                }
+                if (event.getLockedSound() != null) {
+                    event.getPlayer().playSound(event.getLockedSound());
+                }
+                return false;
+            } else {
+                return true;
+            }
+        } else { // logic below is replaced by logic above
+            // Paper end - Add BlockLockCheckEvent
+            if (!p_58630_.isSpectator() && !p_58631_.unlocksWith(p_58630_.getMainHandItem())) {
+                p_58630_.displayClientMessage(Component.translatable("container.isLocked", p_58632_), true); // Paper - diff on change
+                p_58630_.playNotifySound(SoundEvents.CHEST_LOCKED, SoundSource.BLOCKS, 1.0F, 1.0F);
+                return false;
+            } else {
+                return true;
+            }
+        } // Paper - Add BlockLockCheckEvent
     }
 
     protected abstract NonNullList<ItemStack> getItems();
@@ -166,4 +_,12 @@
         p_330762_.remove("Lock");
         p_330762_.remove("Items");
     }
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.Location getLocation() {
+        if (level == null) return null;
+        return new org.bukkit.Location(level.getWorld(), worldPosition.getX(), worldPosition.getY(), worldPosition.getZ());
+    }
+    // CraftBukkit end
 }
