--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -66,10 +_,21 @@
     protected abstract Component getDefaultName();
 
     public boolean canOpen(Player player) {
-        return this.lockKey.canUnlock(player);
+        // SparklyPaper - parallel world ticking (see: PARALLEL_NOTES.md - Opening an inventory after a world switch)
+        if (player instanceof net.minecraft.server.level.ServerPlayer serverPlayer && getLevel() != serverPlayer.level()) {
+            net.minecraft.server.MinecraftServer.LOGGER.warn("Player " + serverPlayer.getScoreboardName() + " (" + serverPlayer.getStringUUID() + ") attempted to open a BlockEntity @ " + this.getLevel().getWorld().getName() + " " + this.getBlockPos().getX() + ", " + this.getBlockPos().getY() + ", " + this.getBlockPos().getZ() + " while they were in a different world " + serverPlayer.level().getWorld().getName() + " than the block themselves!");
+            return false;
+        }
+        // SparklyPaper end
+        return org.bukkit.craftbukkit.event.CraftEventFactory.callBlockLockCheckEvent(this, this.lockKey, this.getDisplayName(), player); // Paper - Call BlockLockCheckEvent
     }
 
     public static void sendChestLockedNotifications(Vec3 pos, Player player, Component displayName) {
+        // Paper start - BlockLockCheckEvent
+        if (org.bukkit.craftbukkit.event.CraftEventFactory.sendChestLockedNotifications(pos)) {
+            return;
+        }
+        // Paper end - BlockLockCheckEvent
         Level level = player.level();
         player.sendOverlayMessage(Component.translatable("container.isLocked", displayName));
         if (!level.isClientSide()) {
@@ -118,9 +_,17 @@
 
     @Override
     public void setItem(int slot, ItemStack itemStack) {
+        setItem(slot, itemStack, false);
+    }
+
+    // Neo: Skip side-effects if insideTransaction is true so the caller can defer them until the transaction commits
+    @Override
+    public void setItem(int slot, ItemStack itemStack, boolean insideTransaction) {
         this.getItems().set(slot, itemStack);
         itemStack.limitSize(this.getMaxStackSize(itemStack));
-        this.setChanged();
+        if (!insideTransaction) {
+            this.setChanged();
+        }
     }
 
     @Override
@@ -170,4 +_,12 @@
         output.discard("lock");
         output.discard("Items");
     }
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.Location getLocation() {
+        if (level == null) return null;
+        return org.bukkit.craftbukkit.util.CraftLocation.toBukkit(this.worldPosition, this.level);
+    }
+    // CraftBukkit end
 }
