--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -64,15 +_,15 @@
     private static final String TAG_SECONDARY = "secondary_effect";
     List<BeaconBlockEntity.BeaconBeamSection> beamSections = Lists.newArrayList();
     private List<BeaconBlockEntity.BeaconBeamSection> checkingBeamSections = Lists.newArrayList();
-    public int levels;
+    int levels;
     private int lastCheckY;
     @Nullable
-    public Holder<MobEffect> primaryPower;
-    @Nullable
-    public Holder<MobEffect> secondaryPower;
-    @Nullable
-    public Component name;
-    public LockCode lockKey = LockCode.NO_LOCK;
+    Holder<MobEffect> primaryPower;
+    @Nullable
+    Holder<MobEffect> secondaryPower;
+    @Nullable
+    private Component name;
+    private LockCode lockKey = LockCode.NO_LOCK;
     private final ContainerData dataAccess = new ContainerData() {
         @Override
         public int get(int p_58711_) {
@@ -137,8 +_,8 @@
 
         for (int i1 = 0; i1 < 10 && blockpos.getY() <= l; i1++) {
             BlockState blockstate = p_155108_.getBlockState(blockpos);
-            if (blockstate.getBlock() instanceof BeaconBeamBlock beaconbeamblock) {
-                int j1 = beaconbeamblock.getColor().getTextureDiffuseColor();
+            Integer j1 = blockstate.getBeaconColorMultiplier(p_155108_, blockpos, p_155109_);
+            if (j1 != null) {
                 if (p_155111_.checkingBeamSections.size() <= 1) {
                     beaconblockentity$beaconbeamsection = new BeaconBlockEntity.BeaconBeamSection(j1);
                     p_155111_.checkingBeamSections.add(beaconblockentity$beaconbeamsection);
