--- a/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
@@ -86,6 +_,7 @@
     public static final int MIN_OCCUPATION_TICKS_NECTARLESS = 600;
     private List<BeehiveBlockEntity.BeeData> stored = Lists.newArrayList();
     public @Nullable BlockPos savedFlowerPos;
+    public int maxBees = 3; // CraftBukkit - allow setting max amount of bees a hive can hold
 
     public BeehiveBlockEntity(BlockPos worldPosition, BlockState blockState) {
         super(BlockEntityType.BEEHIVE, worldPosition, blockState);
