--- a/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
@@ -86,6 +_,7 @@
     public static final int MIN_OCCUPATION_TICKS_NECTARLESS = 600;
     private List<BeehiveBlockEntity.BeeData> stored = Lists.newArrayList();
     public @Nullable BlockPos savedFlowerPos;
+    public int maxBees = 3; // CraftBukkit - allow setting max amount of bees a hive can hold
 
     public BeehiveBlockEntity(BlockPos worldPosition, BlockState blockState) {
         super(BlockEntityType.BEEHIVE, worldPosition, blockState);
@@ -150,6 +_,20 @@
         return spawned;
     }
 
+    public List<Entity> releaseBees(BlockState blockstate, BeehiveBlockEntity.BeeReleaseStatus beehiveblockentity_beereleasestatus, boolean force) {
+        List<Entity> list = Lists.newArrayList();
+
+        this.stored.removeIf((beehiveblockentity_beedata) -> {
+            return releaseOccupant(this.level, this.worldPosition, blockstate, beehiveblockentity_beedata.toOccupant(), list, beehiveblockentity_beereleasestatus, this.savedFlowerPos, force);
+            // CraftBukkit end
+        });
+        if (!list.isEmpty()) {
+            super.setChanged();
+        }
+
+        return list;
+    }
+
     @VisibleForDebug
     public int getOccupantCount() {
         return this.stored.size();
@@ -261,6 +_,10 @@
                 }
             }
         }
+    }
+
+    private static boolean releaseOccupant(Level level, BlockPos blockPos, BlockState state, BeehiveBlockEntity.Occupant beeData, @Nullable List<Entity> spawned, BeehiveBlockEntity.BeeReleaseStatus releaseStatus, @Nullable BlockPos savedFlowerPos, boolean force) {
+        return releaseOccupant(level, blockPos, state, beeData, spawned, releaseStatus, savedFlowerPos);
     }
 
     private boolean hasSavedFlowerPos() {
