--- a/net/minecraft/world/level/block/entity/BellBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BellBlockEntity.java
@@ -32,8 +_,8 @@
     public boolean shaking;
     public Direction clickDirection;
     private List<LivingEntity> nearbyEntities;
-    public boolean resonating;
-    public int resonationTicks;
+    private boolean resonating;
+    private int resonationTicks;
 
     public BellBlockEntity(BlockPos p_155173_, BlockState p_155174_) {
         super(BlockEntityType.BELL, p_155173_, p_155174_);
