--- a/net/minecraft/world/level/block/entity/BlockEntityTypes.java
+++ b/net/minecraft/world/level/block/entity/BlockEntityTypes.java
@@ -215,7 +_,7 @@
     public static final BlockEntityType<PotentSulfurBlockEntity> POTENT_SULFUR = register(
         BlockEntityTypeIds.POTENT_SULFUR, PotentSulfurBlockEntity::new, Blocks.POTENT_SULFUR
     );
-    static final Set<BlockEntityType<?>> OP_ONLY_CUSTOM_DATA = Set.of(COMMAND_BLOCK, LECTERN, SIGN, HANGING_SIGN, MOB_SPAWNER, TRIAL_SPAWNER);
+    static final Set<BlockEntityType<?>> OP_ONLY_CUSTOM_DATA = Set.of(COMMAND_BLOCK, LECTERN, SIGN, HANGING_SIGN, MOB_SPAWNER, TRIAL_SPAWNER, BlockEntityTypes.CHEST); // CraftBukkit
 
     private static <T extends BlockEntity> BlockEntityType<T> register(
         ResourceKey<BlockEntityType<?>> key, BlockEntityType.BlockEntitySupplier<? extends T> factory, Block... validBlocks
