--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.level.block.entity;
 
 import java.util.Arrays;
+import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -8,6 +_,7 @@
 import net.minecraft.core.NonNullList;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.chat.Component;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.world.ContainerHelper;
 import net.minecraft.world.Containers;
 import net.minecraft.world.WorldlyContainer;
@@ -22,6 +_,8 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BrewingStandBlock;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
 
 public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements WorldlyContainer {
     private static final int INGREDIENT_SLOT = 3;
@@ -65,6 +_,37 @@
         }
     };
 
+    // CraftBukkit start - add fields and methods
+    private int lastTick = MinecraftServer.currentTick;
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+
+    public void onOpen(CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    public List<ItemStack> getContents() {
+        return this.items;
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return maxStack;
+    }
+
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+    // CraftBukkit end
+
     public BrewingStandBlockEntity(BlockPos p_155283_, BlockState p_155284_) {
         super(BlockEntityType.BREWING_STAND, p_155283_, p_155284_);
     }
@@ -164,6 +_,7 @@
     }
 
     private static void doBrew(Level p_155291_, BlockPos p_155292_, NonNullList<ItemStack> p_155293_) {
+        if (net.neoforged.neoforge.event.EventHooks.onPotionAttemptBrew(p_155293_)) return;
         ItemStack itemstack = p_155293_.get(3);
         PotionBrewing potionbrewing = p_155291_.potionBrewing();
 
@@ -171,15 +_,17 @@
             p_155293_.set(i, potionbrewing.mix(itemstack, p_155293_.get(i)));
         }
 
-        itemstack.shrink(1);
-        if (itemstack.getItem().hasCraftingRemainingItem()) {
-            ItemStack itemstack1 = new ItemStack(itemstack.getItem().getCraftingRemainingItem());
+        net.neoforged.neoforge.event.EventHooks.onPotionBrewed(p_155293_);
+        if (itemstack.hasCraftingRemainingItem()) {
+            ItemStack itemstack1 = itemstack.getCraftingRemainingItem();
+            itemstack.shrink(1);
             if (itemstack.isEmpty()) {
                 itemstack = itemstack1;
             } else {
                 Containers.dropItemStack(p_155291_, (double)p_155292_.getX(), (double)p_155292_.getY(), (double)p_155292_.getZ(), itemstack1);
             }
         }
+        else itemstack.shrink(1);
 
         p_155293_.set(3, itemstack);
         p_155291_.levelEvent(1035, p_155292_, 0);
@@ -208,13 +_,13 @@
 
     @Override
     public boolean canPlaceItem(int p_59017_, ItemStack p_59018_) {
+        PotionBrewing potionbrewing = this.level != null ? this.level.potionBrewing() : PotionBrewing.EMPTY;
         if (p_59017_ == 3) {
-            PotionBrewing potionbrewing = this.level != null ? this.level.potionBrewing() : PotionBrewing.EMPTY;
             return potionbrewing.isIngredient(p_59018_);
         } else {
             return p_59017_ == 4
                 ? p_59018_.is(Items.BLAZE_POWDER)
-                : (p_59018_.is(Items.POTION) || p_59018_.is(Items.SPLASH_POTION) || p_59018_.is(Items.LINGERING_POTION) || p_59018_.is(Items.GLASS_BOTTLE))
+                : (potionbrewing.isInput(p_59018_) || p_59018_.is(Items.GLASS_BOTTLE))
                     && this.getItem(p_59017_).isEmpty();
         }
     }
