--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.level.block.entity;
 
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -8,6 +_,7 @@
 import net.minecraft.core.NonNullList;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.chat.Component;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.world.ContainerHelper;
 import net.minecraft.world.Containers;
 import net.minecraft.world.WorldlyContainer;
@@ -22,6 +_,16 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BrewingStandBlock;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.event.block.BrewingStartEvent;
+import org.bukkit.event.inventory.BrewEvent;
+import org.bukkit.event.inventory.BrewingStandFuelEvent;
+import org.bukkit.inventory.BrewerInventory;
+import org.bukkit.inventory.InventoryHolder;
 
 public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements WorldlyContainer {
     private static final int INGREDIENT_SLOT = 3;
@@ -35,6 +_,7 @@
     public static final int NUM_DATA_VALUES = 2;
     private NonNullList<ItemStack> items = NonNullList.withSize(5, ItemStack.EMPTY);
     public int brewTime;
+    public int recipeBrewTime = 400; // Paper - Add recipeBrewTime
     private boolean[] lastPotionCount;
     private Item ingredient;
     public int fuel;
@@ -44,6 +_,9 @@
             return switch (p_59038_) {
                 case 0 -> BrewingStandBlockEntity.this.brewTime;
                 case 1 -> BrewingStandBlockEntity.this.fuel;
+                // Paper start - Add recipeBrewTime
+                case 2 -> BrewingStandBlockEntity.this.recipeBrewTime;
+                // Paper end - Add recipeBrewTime
                 default -> 0;
             };
         }
@@ -56,15 +_,50 @@
                     break;
                 case 1:
                     BrewingStandBlockEntity.this.fuel = p_59041_;
+                    // Paper start - Add recipeBrewTime
+                case 2:
+                    BrewingStandBlockEntity.this.recipeBrewTime = p_59041_;
+                    break;
+                // Paper end - Add recipeBrewTime
             }
         }
 
         @Override
         public int getCount() {
-            return 2;
+            return 3; // Paper - Add recipeBrewTime
         }
     };
 
+    // CraftBukkit start - add fields and methods
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+
+    public void onOpen(CraftHumanEntity who) {
+        transaction.add(who);
+    }
+
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    public List<ItemStack> getContents() {
+        return this.items;
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return maxStack;
+    }
+
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+    // CraftBukkit end
+
     public BrewingStandBlockEntity(BlockPos p_155283_, BlockState p_155284_) {
         super(BlockEntityType.BREWING_STAND, p_155283_, p_155284_);
     }
@@ -92,19 +_,33 @@
     public static void serverTick(Level p_155286_, BlockPos p_155287_, BlockState p_155288_, BrewingStandBlockEntity p_155289_) {
         ItemStack itemstack = p_155289_.items.get(4);
         if (p_155289_.fuel <= 0 && itemstack.is(Items.BLAZE_POWDER)) {
-            p_155289_.fuel = 20;
-            itemstack.shrink(1);
+            // CraftBukkit start
+            BrewingStandFuelEvent event = new BrewingStandFuelEvent(CraftBlock.at(p_155286_, p_155287_), CraftItemStack.asCraftMirror(itemstack), 20);
+            p_155286_.getCraftServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return;
+            }
+
+            p_155289_.fuel = event.getFuelPower();
+            if (p_155289_.fuel > 0 && event.isConsuming()) {
+                itemstack.shrink(1);
+            }
+            // CraftBukkit end
             setChanged(p_155286_, p_155287_, p_155288_);
         }
 
         boolean flag = isBrewable(p_155286_.potionBrewing(), p_155289_.items);
         boolean flag1 = p_155289_.brewTime > 0;
         ItemStack itemstack1 = p_155289_.items.get(3);
+
         if (flag1) {
-            p_155289_.brewTime--;
-            boolean flag2 = p_155289_.brewTime == 0;
+            --p_155289_.brewTime; // Paper - remove anti tick skipping measures / wall time - revert to vanilla
+            boolean flag2 = p_155289_.brewTime <= 0; // == -> <=
+            // CraftBukkit end
             if (flag2 && flag) {
-                doBrew(p_155286_, p_155287_, p_155289_.items);
+                brewingStandBlock = p_155289_;
+                doBrew(p_155286_, p_155287_, p_155289_.items); // CraftBukkit
             } else if (!flag || !itemstack1.is(p_155289_.ingredient)) {
                 p_155289_.brewTime = 0;
             }
@@ -112,7 +_,12 @@
             setChanged(p_155286_, p_155287_, p_155288_);
         } else if (flag && p_155289_.fuel > 0) {
             p_155289_.fuel--;
-            p_155289_.brewTime = 400;
+            // CraftBukkit start
+            BrewingStartEvent event = new BrewingStartEvent(CraftBlock.at(p_155286_, p_155287_), CraftItemStack.asCraftMirror(itemstack1), 400);
+            p_155286_.getCraftServer().getPluginManager().callEvent(event);
+            p_155289_.recipeBrewTime = event.getRecipeBrewTime(); // Paper - use recipe brew time from event
+            p_155289_.brewTime = event.getBrewingTime(); // 400 -> event.getTotalBrewTime() // Paper - use brewing time from event
+            // CraftBukkit end
             p_155289_.ingredient = itemstack1.getItem();
             setChanged(p_155286_, p_155287_, p_155288_);
         }
@@ -162,24 +_,48 @@
             return false;
         }
     }
-
+    private static BrewingStandBlockEntity brewingStandBlock;
     private static void doBrew(Level p_155291_, BlockPos p_155292_, NonNullList<ItemStack> p_155293_) {
+        if (net.neoforged.neoforge.event.EventHooks.onPotionAttemptBrew(p_155293_)) return;
         ItemStack itemstack = p_155293_.get(3);
         PotionBrewing potionbrewing = p_155291_.potionBrewing();
 
-        for (int i = 0; i < 3; i++) {
-            p_155293_.set(i, potionbrewing.mix(itemstack, p_155293_.get(i)));
-        }
+        // CraftBukkit start
+        List<org.bukkit.inventory.ItemStack> brewResults = new ArrayList<>(3);
+        for (int i = 0; i < 3; ++i) {
+            var input = p_155293_.get(i);
+            brewResults.add(i, CraftItemStack.asCraftMirror(potionbrewing.mix(itemstack, input)));
+        }
+        BrewingStandBlockEntity brewingStandBlockEntity = brewingStandBlock;
+        InventoryHolder owner = brewingStandBlockEntity == null ? null : brewingStandBlockEntity.getOwner();
+        if (owner != null) {
+            BrewEvent event = new BrewEvent(CraftBlock.at(p_155291_, p_155292_), (BrewerInventory) owner.getInventory(), brewResults, brewingStandBlockEntity.fuel);
+            Bukkit.getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                return;
+            } else {
+                for (int i = 0; i < 3; ++i) {
+                    if (i < brewResults.size()) {
+                        p_155293_.set(i, CraftItemStack.asNMSCopy(brewResults.get(i)));
+                    } else {
+                        p_155293_.set(i, ItemStack.EMPTY);
+                    }
+                }
+            }
+        }
+        // CraftBukkit end
 
-        itemstack.shrink(1);
-        if (itemstack.getItem().hasCraftingRemainingItem()) {
-            ItemStack itemstack1 = new ItemStack(itemstack.getItem().getCraftingRemainingItem());
+        net.neoforged.neoforge.event.EventHooks.onPotionBrewed(p_155293_);
+        if (itemstack.hasCraftingRemainingItem()) {
+            ItemStack itemstack1 = itemstack.getCraftingRemainingItem();
+            itemstack.shrink(1);
             if (itemstack.isEmpty()) {
                 itemstack = itemstack1;
             } else {
                 Containers.dropItemStack(p_155291_, (double)p_155292_.getX(), (double)p_155292_.getY(), (double)p_155292_.getZ(), itemstack1);
             }
         }
+        else itemstack.shrink(1);
 
         p_155293_.set(3, itemstack);
         p_155291_.levelEvent(1035, p_155292_, 0);
@@ -208,13 +_,13 @@
 
     @Override
     public boolean canPlaceItem(int p_59017_, ItemStack p_59018_) {
+        PotionBrewing potionbrewing = this.level != null ? this.level.potionBrewing() : PotionBrewing.EMPTY;
         if (p_59017_ == 3) {
-            PotionBrewing potionbrewing = this.level != null ? this.level.potionBrewing() : PotionBrewing.EMPTY;
             return potionbrewing.isIngredient(p_59018_);
         } else {
             return p_59017_ == 4
                 ? p_59018_.is(Items.BLAZE_POWDER)
-                : (p_59018_.is(Items.POTION) || p_59018_.is(Items.SPLASH_POTION) || p_59018_.is(Items.LINGERING_POTION) || p_59018_.is(Items.GLASS_BOTTLE))
+                : (potionbrewing.isInput(p_59018_) || p_59018_.is(Items.GLASS_BOTTLE) || potionbrewing.isCustomInput(p_59018_))
                     && this.getItem(p_59017_).isEmpty();
         }
     }
