--- a/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
@@ -43,12 +_,12 @@
     private int brushCount;
     private long brushCountResetsAtTick;
     private long coolDownEndsAtTick;
-    public ItemStack item = ItemStack.EMPTY;
+    private ItemStack item = ItemStack.EMPTY;
     @Nullable
     private Direction hitDirection;
     @Nullable
-    public ResourceKey<LootTable> lootTable;
-    public long lootTableSeed;
+    private ResourceKey<LootTable> lootTable;
+    private long lootTableSeed;
 
     public BrushableBlockEntity(BlockPos p_277558_, BlockState p_278093_) {
         super(BlockEntityType.BRUSHABLE_BLOCK, p_277558_, p_278093_);
