--- a/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
@@ -35,6 +_,13 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.util.Arrays;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.block.BlockBrushEvent;
+// CraftBukkit end
+
 public class BrushableBlockEntity extends BlockEntity {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final String LOOT_TABLE_TAG = "LootTable";
@@ -77,7 +_,12 @@
                 if (previousCompletionState != completionState) {
                     BlockState previousState = this.getBlockState();
                     BlockState state = previousState.setValue(BlockStateProperties.DUSTED, completionState);
-                    level.setBlock(this.getBlockPos(), state, 3);
+                    // CraftBukkit start
+                    BlockBrushEvent event = CraftEventFactory.callBlockBrushEvent(level, this.getBlockPos(), state, 3, (ServerPlayer) user);
+                    if (!event.isCancelled()) {
+                        event.getNewState().update(true);
+                    }
+                    // CraftBukkit end
                 }
 
                 return false;
@@ -114,9 +_,10 @@
     }
 
     private void brushingCompleted(ServerLevel level, LivingEntity user, ItemStack brush) {
-        this.dropContent(level, user, brush);
+        // this.dropContent(serverlevel, livingentity, itemstack); // CraftBukkit - moved down
         BlockState blockState = this.getBlockState();
-        level.levelEvent(3008, this.getBlockPos(), Block.getId(blockState));
+
+        // worldserver.levelEvent(3008, this.getBlockPos(), Block.getId(iblockdata)); // CraftBukkit - moved down
         Block turnsInto;
         if (this.getBlockState().getBlock() instanceof BrushableBlock brushableBlock) {
             turnsInto = brushableBlock.getTurnsInto();
@@ -124,7 +_,15 @@
             turnsInto = Blocks.AIR;
         }
 
-        level.setBlock(this.worldPosition, turnsInto.defaultBlockState(), 3);
+        // CraftBukkit start
+        BlockBrushEvent event = CraftEventFactory.callBlockBrushEvent(level, this.worldPosition, turnsInto.defaultBlockState(), 3, (ServerPlayer) user);
+        if (!event.isCancelled()) {
+            this.dropContent(level, user, brush); // CraftBukkit - from above
+            level.levelEvent(3008, this.getBlockPos(), Block.getId(blockState)); // CraftBukkit - from above
+
+            event.getNewState().update(true);
+        }
+        // CraftBukkit end
     }
 
     private void dropContent(ServerLevel level, LivingEntity user, ItemStack brush) {
@@ -140,7 +_,12 @@
             double zo = dropPos.getZ() + 0.5 * centerRange + halfSize;
             ItemEntity entity = new ItemEntity(level, xo, yo, zo, this.item.split(level.getRandom().nextInt(21) + 10));
             entity.setDeltaMovement(Vec3.ZERO);
-            level.addFreshEntity(entity);
+            // CraftBukkit start
+            if (user instanceof ServerPlayer serverplayer) {
+                org.bukkit.block.Block bblock = CraftBlock.at(this.level, this.worldPosition);
+                CraftEventFactory.handleBlockDropItemEvent(bblock, bblock.getState(), serverplayer, Arrays.asList(entity));
+            }
+            // CraftBukkit end
             this.item = ItemStack.EMPTY;
         }
     }
@@ -168,7 +_,7 @@
     }
 
     private boolean tryLoadLootTable(ValueInput input) {
-        this.lootTable = input.read("LootTable", LootTable.KEY_CODEC).orElse(null);
+        this.lootTable = input.read("LootTable", LootTable.KEY_CODEC).orElse(null); // CraftBukkit - decompile error
         this.lootTableSeed = input.getLongOr("LootTableSeed", 0L);
         return this.lootTable != null;
     }
@@ -211,7 +_,7 @@
             this.item = ItemStack.EMPTY;
         }
 
-        this.hitDirection = input.read("hit_direction", Direction.LEGACY_ID_CODEC).orElse(null);
+        this.hitDirection = input.read("hit_direction", Direction.LEGACY_ID_CODEC).orElse(null); // CraftBukkit - decompile error
     }
 
     @Override
