--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -28,6 +_,8 @@
 import net.minecraft.world.level.block.CampfireBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 
 public class CampfireBlockEntity extends BlockEntity implements Clearable {
     private static final int BURN_COOL_SPEED = 2;
@@ -36,11 +_,13 @@
     public final int[] cookingProgress = new int[4];
     public final int[] cookingTime = new int[4];
     private final RecipeManager.CachedCheck<SingleRecipeInput, CampfireCookingRecipe> quickCheck = RecipeManager.createCheck(RecipeType.CAMPFIRE_COOKING);
+    public final boolean[] stopCooking = new boolean[4]; // Paper - Add more Campfire API
 
     public CampfireBlockEntity(BlockPos p_155301_, BlockState p_155302_) {
         super(BlockEntityType.CAMPFIRE, p_155301_, p_155302_);
     }
 
+    static Optional<RecipeHolder<CampfireCookingRecipe>> recipeHolderOptional = Optional.empty();
     public static void cookTick(Level p_155307_, BlockPos p_155308_, BlockState p_155309_, CampfireBlockEntity p_155310_) {
         boolean flag = false;
 
@@ -48,14 +_,34 @@
             ItemStack itemstack = p_155310_.items.get(i);
             if (!itemstack.isEmpty()) {
                 flag = true;
-                p_155310_.cookingProgress[i]++;
+                if (!p_155310_.stopCooking[i]) { // Paper - Add more Campfire API
+                    p_155310_.cookingProgress[i]++;
+                } // Paper - Add more Campfire API
                 if (p_155310_.cookingProgress[i] >= p_155310_.cookingTime[i]) {
                     SingleRecipeInput singlerecipeinput = new SingleRecipeInput(itemstack);
-                    ItemStack itemstack1 = p_155310_.quickCheck
-                        .getRecipeFor(singlerecipeinput, p_155307_)
+                    // Paper start - add recipe to cook events
+                    recipeHolderOptional = p_155310_.quickCheck.getRecipeFor(singlerecipeinput, p_155307_);
+                    ItemStack itemstack1 = recipeHolderOptional
                         .map(p_344662_ -> p_344662_.value().assemble(singlerecipeinput, p_155307_.registryAccess()))
                         .orElse(itemstack);
                     if (itemstack1.isItemEnabled(p_155307_.enabledFeatures())) {
+                        {
+                            // CraftBukkit start - fire BlockCookEvent
+                            CraftItemStack source = CraftItemStack.asCraftMirror(itemstack);
+                            org.bukkit.inventory.ItemStack result = CraftItemStack.asBukkitCopy(itemstack1);
+
+                            org.bukkit.event.block.BlockCookEvent blockCookEvent = new org.bukkit.event.block.BlockCookEvent(CraftBlock.at(p_155307_, p_155308_), source, result, (org.bukkit.inventory.CookingRecipe<?>) recipeHolderOptional.map(RecipeHolder::toBukkitRecipe).orElse(null)); // Paper - Add recipe to cook events
+                            recipeHolderOptional = Optional.empty();
+                            p_155307_.getCraftServer().getPluginManager().callEvent(blockCookEvent);
+
+                            if (blockCookEvent.isCancelled()) {
+                                return;
+                            }
+
+                            result = blockCookEvent.getResult();
+                            itemstack1 = CraftItemStack.asNMSCopy(result);
+                            // CraftBukkit end
+                        }
                         Containers.dropItemStack(p_155307_, (double)p_155308_.getX(), (double)p_155308_.getY(), (double)p_155308_.getZ(), itemstack1);
                         p_155310_.items.set(i, ItemStack.EMPTY);
                         p_155307_.sendBlockUpdated(p_155308_, p_155309_, p_155309_, 3);
@@ -134,6 +_,17 @@
             int[] aint1 = p_155312_.getIntArray("CookingTotalTimes");
             System.arraycopy(aint1, 0, this.cookingTime, 0, Math.min(this.cookingTime.length, aint1.length));
         }
+
+        // Paper start - Add more Campfire API
+        if (p_155312_.contains("Paper.StopCooking", org.bukkit.craftbukkit.util.CraftMagicNumbers.NBT.TAG_BYTE_ARRAY)) {
+            byte[] abyte = p_155312_.getByteArray("Paper.StopCooking");
+            boolean[] cookingState = new boolean[4];
+            for (int index = 0; index < abyte.length; index++) {
+                cookingState[index] = abyte[index] == 1;
+            }
+            System.arraycopy(cookingState, 0, this.stopCooking, 0, Math.min(this.stopCooking.length, abyte.length));
+        }
+        // Paper end - Add more Campfire API
     }
 
     @Override
@@ -142,6 +_,13 @@
         ContainerHelper.saveAllItems(p_187486_, this.items, true, p_324005_);
         p_187486_.putIntArray("CookingTimes", this.cookingProgress);
         p_187486_.putIntArray("CookingTotalTimes", this.cookingTime);
+        // Paper start - Add more Campfire API
+        byte[] cookingState = new byte[4];
+        for (int index = 0; index < cookingState.length; index++) {
+            cookingState[index] = (byte) (this.stopCooking[index] ? 1 : 0);
+        }
+        p_187486_.putByteArray("Paper.StopCooking", cookingState);
+        // Paper end - Add more Campfire API
     }
 
     public ClientboundBlockEntityDataPacket getUpdatePacket() {
@@ -163,7 +_,11 @@
         for (int i = 0; i < this.items.size(); i++) {
             ItemStack itemstack = this.items.get(i);
             if (itemstack.isEmpty()) {
-                this.cookingTime[i] = p_238287_;
+                // CraftBukkit start
+                org.bukkit.event.block.CampfireStartEvent event = new org.bukkit.event.block.CampfireStartEvent(CraftBlock.at(this.level,this.worldPosition), CraftItemStack.asCraftMirror(p_238286_), (org.bukkit.inventory.CampfireRecipe) getCookableRecipe(p_238286_).get().toBukkitRecipe());
+                this.level.getCraftServer().getPluginManager().callEvent(event);
+                this.cookingTime[i] = event.getTotalCookTime(); // i -> event.getTotalCookTime()
+                // CraftBukkit end
                 this.cookingProgress[i] = 0;
                 this.items.set(i, p_238286_.consumeAndReturn(1, p_347582_));
                 this.level.gameEvent(GameEvent.BLOCK_CHANGE, this.getBlockPos(), GameEvent.Context.of(p_347582_, this.getBlockState()));
