--- a/net/minecraft/world/level/block/entity/ChestBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ChestBlockEntity.java
@@ -27,7 +_,7 @@
 public class ChestBlockEntity extends RandomizableContainerBlockEntity implements LidBlockEntity {
     private static final int EVENT_SET_OPEN_COUNT = 1;
     private NonNullList<ItemStack> items = NonNullList.withSize(27, ItemStack.EMPTY);
-    public final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
+    private final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
         @Override
         protected void onOpen(Level p_155357_, BlockPos p_155358_, BlockState p_155359_) {
             ChestBlockEntity.playSound(p_155357_, p_155358_, p_155359_, SoundEvents.CHEST_OPEN);
@@ -95,7 +_,7 @@
         p_155347_.chestLidController.tickLid();
     }
 
-    public static void playSound(Level p_155339_, BlockPos p_155340_, BlockState p_155341_, SoundEvent p_155342_) {
+    static void playSound(Level p_155339_, BlockPos p_155340_, BlockState p_155341_, SoundEvent p_155342_) {
         ChestType chesttype = p_155341_.getValue(ChestBlock.TYPE);
         if (chesttype != ChestType.LEFT) {
             double d0 = (double)p_155340_.getX() + 0.5;
@@ -171,6 +_,17 @@
     @Override
     protected AbstractContainerMenu createMenu(int p_59082_, Inventory p_59083_) {
         return ChestMenu.threeRows(p_59082_, p_59083_, this);
+    }
+
+    @Override
+    public void setBlockState(BlockState p_155251_) {
+        var oldState = getBlockState();
+        super.setBlockState(p_155251_);
+        // Neo: Chest state change might change the chest item handler -> invalidate
+        if ((oldState.getValue(ChestBlock.FACING) != p_155251_.getValue(ChestBlock.FACING))
+                || (oldState.getValue(ChestBlock.TYPE) != p_155251_.getValue(ChestBlock.TYPE))) {
+            this.invalidateCapabilities();
+        }
     }
 
     public void recheckOpen() {
