--- a/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
@@ -25,7 +_,7 @@
     public static final int MAX_BOOKS_IN_STORAGE = 6;
     private static final Logger LOGGER = LogUtils.getLogger();
     private final NonNullList<ItemStack> items = NonNullList.withSize(6, ItemStack.EMPTY);
-    public int lastInteractedSlot = -1;
+    private int lastInteractedSlot = -1;
 
     public ChiseledBookShelfBlockEntity(BlockPos p_249541_, BlockState p_251752_) {
         super(BlockEntityType.CHISELED_BOOKSHELF, p_249541_, p_251752_);
