--- a/net/minecraft/world/level/block/entity/CommandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CommandBlockEntity.java
@@ -15,12 +_,20 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.command.CraftBlockCommandSender;
 
 public class CommandBlockEntity extends BlockEntity {
     private boolean powered;
     private boolean auto;
     private boolean conditionMet;
     private final BaseCommandBlock commandBlock = new BaseCommandBlock() {
+        // CraftBukkit start
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+            return new CraftBlockCommandSender(wrapper, CommandBlockEntity.this);
+        }
+        // CraftBukkit end
+
         @Override
         public void setCommand(String p_59157_) {
             super.setCommand(p_59157_);
@@ -51,7 +_,7 @@
                 Vec3.atCenterOf(CommandBlockEntity.this.worldPosition),
                 new Vec2(0.0F, direction.toYRot()),
                 this.getLevel(),
-                2,
+                this.getLevel().paperConfig().commandBlocks.permissionsLevel,
                 this.getName().getString(),
                 this.getName(),
                 this.getLevel().getServer(),
