--- a/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
@@ -149,10 +_,8 @@
                         BlockPos testPos = worldPosition.offset(ox, oy, ozx);
                         BlockState testBlock = level.getBlockState(testPos);
 
-                        for (Block type : VALID_BLOCKS) {
-                            if (testBlock.is(type)) {
-                                effectBlocks.add(testPos);
-                            }
+                        if (testBlock.isConduitFrame(level, testPos, worldPosition)) {
+                            effectBlocks.add(testPos);
                         }
                     }
                 }
@@ -179,14 +_,32 @@
         }
     }
 
+    public static int getRange(List<BlockPos> effectBlocks) {
+        // CraftBukkit end
+        int i = effectBlocks.size();
+        int j = i / 7 * 16;
+        // CraftBukkit start
+        return j;
+    }
+
+
     private static void updateAndAttackTarget(ServerLevel level, BlockPos worldPosition, BlockState blockState, ConduitBlockEntity entity, boolean isActive) {
+        // CraftBukkit start - add "damageTarget" boolean
+        updateAndAttackTarget(level, worldPosition, blockState, entity, isActive, true);
+    }
+
+    public static void updateAndAttackTarget(ServerLevel level, BlockPos worldPosition, BlockState blockState, ConduitBlockEntity entity, boolean isActive, boolean damageTarget) {
+        // CraftBukkit end
         EntityReference<LivingEntity> newDestroyTarget = updateDestroyTarget(entity.destroyTarget, level, worldPosition, isActive);
         LivingEntity targetEntity = EntityReference.getLivingEntity(newDestroyTarget, level);
-        if (targetEntity != null) {
+
+        // CraftBukkit start
+        if (damageTarget && targetEntity != null) {
             level.playSound(
                 null, targetEntity.getX(), targetEntity.getY(), targetEntity.getZ(), SoundEvents.CONDUIT_ATTACK_TARGET, SoundSource.BLOCKS, 1.0F, 1.0F
             );
             targetEntity.hurtServer(level, level.damageSources().magic(), 4.0F);
+            // CraftBukkit end
         }
 
         if (!Objects.equals(newDestroyTarget, entity.destroyTarget)) {
