--- a/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
@@ -22,6 +_,7 @@
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySpawnReason;
+import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EntityTypes;
 import net.minecraft.world.entity.monster.creaking.Creaking;
 import net.minecraft.world.entity.player.Player;
@@ -108,8 +_,8 @@
 
                 if (entity.creakingInfo == null) {
                     if (updatedState.getValue(CreakingHeartBlock.STATE) == CreakingHeartState.AWAKE) {
-                        if (serverLevel.isSpawningMonsters() && serverLevel.getLevelData().getDifficulty() != Difficulty.PEACEFUL) {
-                            Player player = level.getNearestPlayer(pos.getX(), pos.getY(), pos.getZ(), 32.0, false);
+                        if (serverLevel.isSpawningMonsters()) {
+                            Player player = level.getNearestPlayerThatAffectsSpawning(pos.getX(), pos.getY(), pos.getZ(), 32.0, false); // Paper - Affects Spawning API
                             if (player != null) {
                                 Creaking creaking = spawnProtector(serverLevel, entity);
                                 if (creaking != null) {
@@ -150,7 +_,7 @@
         return this.getCreakingProtector().map(creaking -> Math.sqrt(creaking.distanceToSqr(Vec3.atBottomCenterOf(this.getBlockPos())))).orElse(0.0);
     }
 
-    public void clearCreakingInfo() {
+    private void clearCreakingInfo() {
         this.creakingInfo = null;
         this.setChanged();
     }
@@ -166,7 +_,7 @@
         this.setChanged();
     }
 
-    public Optional<Creaking> getCreakingProtector() {
+    private Optional<Creaking> getCreakingProtector() {
         if (this.creakingInfo == null) {
             return NO_CREAKING;
         }
@@ -197,7 +_,7 @@
         }
     }
 
-    public static @Nullable Creaking spawnProtector(ServerLevel level, CreakingHeartBlockEntity entity) {
+    private static @Nullable Creaking spawnProtector(ServerLevel level, CreakingHeartBlockEntity entity) {
         BlockPos pos = entity.getBlockPos();
         Optional<Creaking> spawnedMob = SpawnUtil.trySpawnMob(
             EntityTypes.CREAKING, EntitySpawnReason.SPAWNER, level, pos, 5, 16, 8, SpawnUtil.Strategy.ON_TOP_OF_COLLIDER_NO_LEAVES, true
@@ -246,7 +_,7 @@
         }
     }
 
-    public Optional<BlockPos> spreadResin(ServerLevel level) {
+    private Optional<BlockPos> spreadResin(ServerLevel level) {
         RandomSource random = level.getRandom();
         Mutable<BlockPos> placedResin = new MutableObject<>(null);
         BlockPos.breadthFirstTraversal(this.worldPosition, 2, 64, (pos, acceptor) -> {
