--- a/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/DecoratedPotBlockEntity.java
@@ -26,7 +_,7 @@
     public long wobbleStartedAtTick;
     @Nullable
     public DecoratedPotBlockEntity.WobbleStyle lastWobbleStyle;
-    public PotDecorations decorations;
+    private PotDecorations decorations;
     private ItemStack item = ItemStack.EMPTY;
     @Nullable
     protected ResourceKey<LootTable> lootTable;
