--- a/net/minecraft/world/level/block/entity/DecoratedPotPatterns.java
+++ b/net/minecraft/world/level/block/entity/DecoratedPotPatterns.java
@@ -3,12 +_,12 @@
 import com.mojang.serialization.Codec;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.data.worldgen.BootstrapContext;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.resources.Identifier;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.resources.ResourceKey;
 
 public class DecoratedPotPatterns {
@@ -35,7 +_,7 @@
     public static final ResourceKey<DecoratedPotPattern> SHELTER = create("shelter");
     public static final ResourceKey<DecoratedPotPattern> SKULL = create("skull");
     public static final ResourceKey<DecoratedPotPattern> SNORT = create("snort");
-    public static final Codec<Holder<DecoratedPotPattern>> CODEC = RegistryCodecs.holder(Registries.DECORATED_POT_PATTERN);
+    public static final Codec<Holder<DecoratedPotPattern>> CODEC = RegistryFixedCodec.create(Registries.DECORATED_POT_PATTERN);
     public static final StreamCodec<RegistryFriendlyByteBuf, Holder<DecoratedPotPattern>> STREAM_CODEC = ByteBufCodecs.holderRegistry(
         Registries.DECORATED_POT_PATTERN
     );
