--- a/net/minecraft/world/level/block/entity/PotDecorations.java
+++ b/net/minecraft/world/level/block/entity/PotDecorations.java
@@ -5,15 +_,12 @@
 import java.util.Optional;
 import java.util.function.Consumer;
 import net.minecraft.ChatFormatting;
-import net.minecraft.core.HolderGetter;
 import net.minecraft.core.component.DataComponentGetter;
-import net.minecraft.core.component.DataComponentPatch;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.chat.CommonComponents;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
-import net.minecraft.references.ItemIds;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStackTemplate;
 import net.minecraft.world.item.TooltipFlag;
@@ -43,11 +_,6 @@
         PotDecorations::front,
         PotDecorations::new
     );
-
-    public static PotDecorations allBrick(HolderGetter<Item> items) {
-        Optional<ItemStackTemplate> brick = Optional.of(new ItemStackTemplate(items.getOrThrow(ItemIds.BRICK), 1, DataComponentPatch.EMPTY));
-        return new PotDecorations(brick, brick, brick, brick);
-    }
 
     @Override
     public void addToTooltip(Item.TooltipContext context, Consumer<Component> consumer, TooltipFlag flag, DataComponentGetter components) {
