--- a/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
@@ -17,8 +_,8 @@
 
 public abstract class RandomizableContainerBlockEntity extends BaseContainerBlockEntity implements RandomizableContainer {
     @Nullable
-    public ResourceKey<LootTable> lootTable;
-    public long lootTableSeed = 0L;
+    protected ResourceKey<LootTable> lootTable;
+    protected long lootTableSeed = 0L;
 
     protected RandomizableContainerBlockEntity(BlockEntityType<?> p_155629_, BlockPos p_155630_, BlockState p_155631_) {
         super(p_155629_, p_155630_, p_155631_);
