--- a/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
@@ -18,8 +_,8 @@
 import org.jspecify.annotations.Nullable;
 
 public abstract class RandomizableContainerBlockEntity extends BaseContainerBlockEntity implements RandomizableContainer {
-    protected @Nullable ResourceKey<LootTable> lootTable;
-    protected long lootTableSeed = 0L;
+    public @Nullable ResourceKey<LootTable> lootTable;
+    public long lootTableSeed = 0L;
 
     protected RandomizableContainerBlockEntity(BlockEntityType<?> type, BlockPos worldPosition, BlockState blockState) {
         super(type, worldPosition, blockState);
@@ -53,7 +_,7 @@
 
     @Override
     public ItemStack getItem(int slot) {
-        this.unpackLootTable(null);
+        if (slot == 0) this.unpackLootTable(null); // Paper - Perf: Optimize Hoppers
         return super.getItem(slot);
     }
 
@@ -118,4 +_,13 @@
         output.discard("LootTable");
         output.discard("LootTableSeed");
     }
+
+    // Paper start - LootTable API
+    final com.destroystokyo.paper.loottable.PaperLootableInventoryData lootableData = new com.destroystokyo.paper.loottable.PaperLootableInventoryData(); // Paper
+
+    @Override
+    public com.destroystokyo.paper.loottable.PaperLootableInventoryData lootableData() {
+        return this.lootableData;
+    }
+    // Paper end - LootTable API
 }
