--- a/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/RandomizableContainerBlockEntity.java
@@ -53,7 +_,7 @@
 
     @Override
     public ItemStack getItem(int slot) {
-        this.unpackLootTable(null);
+        if (slot == 0) this.unpackLootTable(null); // Paper - Perf: Optimize Hoppers
         return super.getItem(slot);
     }
 
