--- a/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -20,6 +_,7 @@
 import net.minecraft.network.chat.Style;
 import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.network.FilteredText;
 import net.minecraft.sounds.SoundEvent;
 import net.minecraft.sounds.SoundEvents;
@@ -30,9 +_,13 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.block.sign.Side;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.block.CraftSign;
+import org.bukkit.craftbukkit.command.CraftBlockCommandSender;
 import org.slf4j.Logger;
 
-public class SignBlockEntity extends BlockEntity {
+public class SignBlockEntity extends BlockEntity implements CommandSource { // CraftBukkit - implements
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int MAX_TEXT_LINE_WIDTH = 90;
     private static final int TEXT_LINE_HEIGHT = 10;
@@ -55,10 +_,15 @@
     }
 
     public boolean isFacingFrontText(Player p_277382_) {
+        // Paper start - More Sign Block API
+        return this.isFacingFrontText(p_277382_.getX(), p_277382_.getZ());
+    }
+    public boolean isFacingFrontText(double x, double z) {
+        // Paper end - More Sign Block API
         if (this.getBlockState().getBlock() instanceof SignBlock signblock) {
             Vec3 vec3 = signblock.getSignHitboxCenterPosition(this.getBlockState());
-            double d0 = p_277382_.getX() - ((double)this.getBlockPos().getX() + vec3.x);
-            double d1 = p_277382_.getZ() - ((double)this.getBlockPos().getZ() + vec3.z);
+            double d0 = x - ((double) this.getBlockPos().getX() + vec3.x); // Paper - More Sign Block API
+            double d1 = z - ((double) this.getBlockPos().getZ() + vec3.z); // Paper - More Sign Block API
             float f = signblock.getYRotationDegrees(this.getBlockState());
             float f1 = (float)(Mth.atan2(d1, d0) * 180.0F / (float)Math.PI) - 90.0F;
             return Mth.degreesDifferenceAbs(f, f1) <= 90.0F;
@@ -146,11 +_,13 @@
 
     public void updateSignText(Player p_278048_, boolean p_278103_, List<FilteredText> p_277990_) {
         if (!this.isWaxed() && p_278048_.getUUID().equals(this.getPlayerWhoMayEdit()) && this.level != null) {
-            this.updateText(p_277776_ -> this.setMessages(p_278048_, p_277990_, p_277776_), p_278103_);
+            this.updateText(p_277776_ -> this.setMessages(p_278048_, p_277990_, p_277776_, p_278103_), p_278103_); // CraftBukkit
             this.setAllowedPlayerEditor(null);
             this.level.sendBlockUpdated(this.getBlockPos(), this.getBlockState(), this.getBlockState(), 3);
         } else {
             LOGGER.warn("Player {} just tried to change non-editable sign", p_278048_.getName().getString());
+            if (p_278048_.distanceToSqr(this.getBlockPos().getX(), this.getBlockPos().getY(), this.getBlockPos().getZ()) < 32 * 32) // Paper - Dont send far away sign update
+            ((ServerPlayer) p_278048_).connection.send(this.getUpdatePacket()); // CraftBukkit
         }
     }
 
@@ -159,18 +_,41 @@
         return this.setText(p_277877_.apply(signtext), p_277426_);
     }
 
-    private SignText setMessages(Player p_277396_, List<FilteredText> p_277744_, SignText p_277359_) {
+    private SignText setMessages(Player p_277396_, List<FilteredText> p_277744_, SignText p_277359_, boolean front) {  // CraftBukkit
+        SignText originalText = p_277359_; // CraftBukkit
         for (int i = 0; i < p_277744_.size(); i++) {
             FilteredText filteredtext = p_277744_.get(i);
             Style style = p_277359_.getMessage(i, p_277396_.isTextFilteringEnabled()).getStyle();
             if (p_277396_.isTextFilteringEnabled()) {
-                p_277359_ = p_277359_.setMessage(i, Component.literal(filteredtext.filteredOrEmpty()).setStyle(style));
+                p_277359_ = p_277359_.setMessage(i, Component.literal(net.minecraft.util.StringUtil.filterText(filteredtext.filteredOrEmpty())).setStyle(style)); // Paper - filter sign text to chat only
             } else {
                 p_277359_ = p_277359_.setMessage(
-                    i, Component.literal(filteredtext.raw()).setStyle(style), Component.literal(filteredtext.filteredOrEmpty()).setStyle(style)
-                );
-            }
-        }
+                    i, Component.literal(net.minecraft.util.StringUtil.filterText(filteredtext.raw())).setStyle(style), Component.literal(net.minecraft.util.StringUtil.filterText(filteredtext.filteredOrEmpty())).setStyle(style)); // Paper - filter sign text to chat only
+            }
+        }
+
+        // CraftBukkit start
+        org.bukkit.entity.Player player = ((net.minecraft.server.level.ServerPlayer) p_277396_).getBukkitEntity();
+        List<net.kyori.adventure.text.Component> lines = new java.util.ArrayList<>(); // Paper - adventure
+
+        for (int i = 0; i < p_277744_.size(); ++i) {
+            lines.add(io.papermc.paper.adventure.PaperAdventure.asAdventure(p_277359_.getMessage(i, p_277396_.isTextFilteringEnabled()))); // Paper - Adventure
+        }
+
+        org.bukkit.event.block.SignChangeEvent event = new org.bukkit.event.block.SignChangeEvent(CraftBlock.at(this.level, this.worldPosition), player, new java.util.ArrayList<>(lines), (front) ? org.bukkit.block.sign.Side.FRONT : org.bukkit.block.sign.Side.BACK);
+        p_277396_.level().getCraftServer().getPluginManager().callEvent(event);
+
+        if (event.isCancelled()) {
+            return originalText;
+        }
+
+        Component[] components = CraftSign.sanitizeLines(event.lines());  // Paper - Adventure
+        for (int i = 0; i < components.length; i++) {
+            if (!java.util.Objects.equals(lines.get(i), event.line(i))) {  // Paper - Adventure
+                p_277359_ = p_277359_.setMessage(i, components[i]);
+            }
+        }
+        // CraftBukkit end
 
         return p_277359_;
     }
@@ -210,7 +_,18 @@
             Style style = component.getStyle();
             ClickEvent clickevent = style.getClickEvent();
             if (clickevent != null && clickevent.getAction() == ClickEvent.Action.RUN_COMMAND) {
-                p_279304_.getServer().getCommands().performPrefixedCommand(createCommandSourceStack(p_279304_, p_279201_, p_278282_), clickevent.getValue());
+                // Paper start - Fix commands from signs not firing command events
+                String command = clickevent.getValue().startsWith("/") ? clickevent.getValue() : "/" + clickevent.getValue();
+                if (org.spigotmc.SpigotConfig.logCommands)  {
+                    LOGGER.info("{} issued server command: {}", p_279304_.getScoreboardName(), command);
+                }
+                io.papermc.paper.event.player.PlayerSignCommandPreprocessEvent event = new io.papermc.paper.event.player.PlayerSignCommandPreprocessEvent((org.bukkit.entity.Player) p_279304_.getBukkitEntity(), command, new org.bukkit.craftbukkit.util.LazyPlayerSet(p_279304_.getServer()), (org.bukkit.block.Sign) CraftBlock.at(this.level, this.worldPosition).getState(), p_278254_ ? Side.FRONT : Side.BACK);
+                if (!event.callEvent()) {
+                    return false;
+                }
+                Player player = ((org.bukkit.craftbukkit.entity.CraftPlayer) event.getPlayer()).getHandle();
+                p_279304_.getServer().getCommands().performPrefixedCommand(createCommandSourceStack(player, p_279201_, p_278282_).withSource(commandSource(player, p_279201_)), event.getMessage());
+                // Paper end - Fix commands from signs not firing command events
                 flag = true;
             }
         }
@@ -218,6 +_,48 @@
         return flag;
     }
 
+    // CraftBukkit start
+    @Override
+    public void sendSystemMessage(Component ichatbasecomponent) {}
+
+    @Override
+    public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+        return wrapper.getEntity() != null ? wrapper.getEntity().getBukkitSender(wrapper) : new CraftBlockCommandSender(wrapper, this);
+    }
+
+    @Override
+    public boolean acceptsSuccess() {
+        return false;
+    }
+
+    @Override
+    public boolean acceptsFailure() {
+        return false;
+    }
+
+    @Override
+    public boolean shouldInformAdmins() {
+        return false;
+    }
+
+    public CommandSource commandSource(@Nullable Player p_279428_, Level p_279359_){
+        // Paper start - Fix commands from signs not firing command events
+        return p_279359_.paperConfig().misc.showSignClickCommandFailureMsgsToPlayer ? new io.papermc.paper.commands.DelegatingCommandSource(SignBlockEntity.this) {
+            @Override
+            public void sendSystemMessage(Component message) {
+                if (p_279428_ != null) {
+                    p_279428_.sendSystemMessage(message);
+                }
+            }
+
+            @Override
+            public boolean acceptsFailure() {
+                return true;
+            }
+        } : this;
+        // Paper end - Fix commands from signs not firing command events
+    }
+
     private static CommandSourceStack createCommandSourceStack(@Nullable Player p_279428_, Level p_279359_, BlockPos p_279430_) {
         String s = p_279428_ == null ? "Sign" : p_279428_.getName().getString();
         Component component = (Component)(p_279428_ == null ? Component.literal("Sign") : p_279428_.getDisplayName());
@@ -246,12 +_,17 @@
 
     @Nullable
     public UUID getPlayerWhoMayEdit() {
+        // CraftBukkit start - unnecessary sign ticking removed, so do this lazily
+        if (this.level != null && this.playerWhoMayEdit != null) {
+            clearInvalidPlayerWhoMayEdit(this, this.level, this.playerWhoMayEdit);
+        }
+        // CraftBukkit end
         return this.playerWhoMayEdit;
     }
 
     private void markUpdated() {
         this.setChanged();
-        this.level.sendBlockUpdated(this.getBlockPos(), this.getBlockState(), this.getBlockState(), 3);
+        if (this.level != null) this.level.sendBlockUpdated(this.getBlockPos(), this.getBlockState(), this.getBlockState(), 3); // CraftBukkit - skip notify if world is null (SPIGOT-5122)
     }
 
     public boolean isWaxed() {
