--- a/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -17,6 +_,7 @@
 import net.minecraft.network.chat.Style;
 import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.network.FilteredText;
 import net.minecraft.server.permissions.LevelBasedPermissionSet;
 import net.minecraft.sounds.SoundEvent;
@@ -33,6 +_,14 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.util.Objects;
+import org.bukkit.block.sign.Side;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.util.CraftChatMessage;
+import org.bukkit.event.block.SignChangeEvent;
+// CraftBukkit end
+
 public class SignBlockEntity extends BlockEntity {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int MAX_TEXT_LINE_WIDTH = 90;
@@ -58,10 +_,15 @@
     }
 
     public boolean isFacingFrontText(Player player) {
+        // Paper start - More Sign Block API
+        return this.isFacingFrontText(player.getX(), player.getZ());
+    }
+    public boolean isFacingFrontText(final double x, final double z) {
+        // Paper end - More Sign Block API
         if (this.getBlockState().getBlock() instanceof SignBlock sign) {
             Vec3 signPositionOffset = sign.getSignHitboxCenterPosition(this.getBlockState());
-            double xd = player.getX() - (this.getBlockPos().getX() + signPositionOffset.x);
-            double zd = player.getZ() - (this.getBlockPos().getZ() + signPositionOffset.z);
+            double xd = x - (this.getBlockPos().getX() + signPositionOffset.x); // Paper - More Sign Block API
+            double zd = z - (this.getBlockPos().getZ() + signPositionOffset.z); // Paper - More Sign Block API
             float signYRot = sign.getYRotationDegrees(this.getBlockState());
             float playerYRot = (float)(Mth.atan2(zd, xd) * 180.0F / (float)Math.PI) - 90.0F;
             return Mth.degreesDifferenceAbs(signYRot, playerYRot) <= 90.0F;
@@ -129,11 +_,12 @@
 
     public void updateSignText(Player player, boolean frontText, List<FilteredText> lines) {
         if (!this.isWaxed() && player.getUUID().equals(this.getPlayerWhoMayEdit()) && this.level != null) {
-            this.updateText(text -> this.setMessages(player, lines, text), frontText);
+            this.updateText(text -> this.setMessages(player, lines, text, frontText), frontText); // CraftBukkit
             this.setAllowedPlayerEditor(null);
             this.level.sendBlockUpdated(this.getBlockPos(), this.getBlockState(), this.getBlockState(), 3);
         } else {
             LOGGER.warn("Player {} just tried to change non-editable sign", player.getPlainTextName());
+            ((ServerPlayer) player).connection.send(this.getUpdatePacket()); // CraftBukkit
         }
     }
 
@@ -142,7 +_,8 @@
         return this.setText(function.apply(text), isFrontText);
     }
 
-    private SignText setMessages(Player player, List<FilteredText> lines, SignText text) {
+    private SignText setMessages(Player player, List<FilteredText> lines, SignText text, boolean front) { // CraftBukkit
+        SignText originalText = text; // CraftBukkit
         for (int i = 0; i < lines.size(); i++) {
             FilteredText line = lines.get(i);
             Style currentTextStyle = text.getMessage(i, player.isTextFilteringEnabled()).getStyle();
@@ -155,6 +_,29 @@
             }
         }
 
+        // CraftBukkit start
+        org.bukkit.entity.Player bukkitPlayer = ((ServerPlayer) player).getBukkitEntity();
+        String[] linesArray = new String[4];
+
+        for (int i = 0; i < lines.size(); ++i) {
+            linesArray[i] = CraftChatMessage.fromComponent(text.getMessage(i, player.isTextFilteringEnabled()));
+        }
+
+        SignChangeEvent event = new SignChangeEvent(CraftBlock.at(this.level, this.worldPosition), bukkitPlayer, linesArray.clone(), (front) ? Side.FRONT : Side.BACK);
+        player.level().getCraftServer().getPluginManager().callEvent(event);
+
+        if (event.isCancelled()) {
+            return originalText;
+        }
+
+        Component[] components = org.bukkit.craftbukkit.block.CraftSign.sanitizeLines(event.getLines());
+        for (int i = 0; i < components.length; i++) {
+            if (!Objects.equals(linesArray[i], event.getLine(i))) {
+                text = text.setMessage(i, components[i]);
+            }
+        }
+        // CraftBukkit end
+
         return text;
     }
 
@@ -202,7 +_,7 @@
                     hasAnyClickCommand = true;
                     break;
                 case ClickEvent.Custom custom:
-                    level.getServer().handleCustomClickAction(custom.id(), custom.payload());
+                    level.getServer().handleCustomClickAction(custom.id(), custom.payload(), (net.minecraft.server.level.ServerPlayer) player, player.getGameProfile());
                     hasAnyClickCommand = true;
                     break;
                 case null:
@@ -213,11 +_,40 @@
         return hasAnyClickCommand;
     }
 
-    private static CommandSourceStack createCommandSourceStack(@Nullable Player player, ServerLevel level, BlockPos pos) {
+    // CraftBukkit start
+    private final CommandSource commandSource = new CommandSource() {
+
+        @Override
+        public void sendSystemMessage(Component component) {}
+
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+            return wrapper.getEntity() != null ? wrapper.getEntity().getBukkitEntity() : new org.bukkit.craftbukkit.command.CraftBlockCommandSender(wrapper, SignBlockEntity.this);
+        }
+
+        @Override
+        public boolean acceptsSuccess() {
+            return false;
+        }
+
+        @Override
+        public boolean acceptsFailure() {
+            return false;
+        }
+
+        @Override
+        public boolean shouldInformAdmins() {
+            return false;
+        }
+    };
+
+    private CommandSourceStack createCommandSourceStack(@Nullable Player player, ServerLevel level, BlockPos pos) {
+        // CraftBukkit end
         String textName = player == null ? "Sign" : player.getPlainTextName();
-        Component displayName = player == null ? Component.literal("Sign") : player.getDisplayName();
+        Component displayName = (Component)(player == null ? Component.literal("Sign") : player.getDisplayName());
+        // CraftBukkit - commandSource
         return new CommandSourceStack(
-            CommandSource.NULL, Vec3.atCenterOf(pos), Vec2.ZERO, level, LevelBasedPermissionSet.GAMEMASTER, textName, displayName, level.getServer(), player
+            commandSource, Vec3.atCenterOf(pos), Vec2.ZERO, level, LevelBasedPermissionSet.GAMEMASTER, textName, displayName, level.getServer(), player
         );
     }
 
@@ -235,12 +_,17 @@
     }
 
     public @Nullable UUID getPlayerWhoMayEdit() {
+        // CraftBukkit start - unnecessary sign ticking removed, so do this lazily
+        if (this.level != null && this.playerWhoMayEdit != null) {
+            clearInvalidPlayerWhoMayEdit(this, this.level, this.playerWhoMayEdit);
+        }
+        // CraftBukkit end
         return this.playerWhoMayEdit;
     }
 
     private void markUpdated() {
         this.setChanged();
-        this.level.sendBlockUpdated(this.getBlockPos(), this.getBlockState(), this.getBlockState(), 3);
+        if (this.level != null) this.level.sendBlockUpdated(this.getBlockPos(), this.getBlockState(), this.getBlockState(), 3); // CraftBukkit - skip notify if world is null (SPIGOT-5122)
     }
 
     public boolean isWaxed() {
