--- a/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -17,6 +_,7 @@
 import net.minecraft.network.chat.Style;
 import net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.network.FilteredText;
 import net.minecraft.server.permissions.LevelBasedPermissionSet;
 import net.minecraft.sounds.SoundEvent;
@@ -201,7 +_,7 @@
                     hasAnyClickCommand = true;
                     break;
                 case ClickEvent.Custom custom:
-                    level.getServer().handleCustomClickAction(custom.id(), custom.payload());
+                    level.getServer().handleCustomClickAction(custom.id(), custom.payload(), (ServerPlayer) player); // CraftBukkit - add player
                     hasAnyClickCommand = true;
                     break;
                 case null:
