--- a/net/minecraft/world/level/block/entity/SignBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SignBlockEntity.java
@@ -37,7 +_,7 @@
     private static final int MAX_TEXT_LINE_WIDTH = 90;
     private static final int TEXT_LINE_HEIGHT = 10;
     @Nullable
-    public UUID playerWhoMayEdit;
+    private UUID playerWhoMayEdit;
     private SignText frontText = this.createDefaultSignText();
     private SignText backText = this.createDefaultSignText();
     private boolean isWaxed;
