--- a/net/minecraft/world/level/block/entity/SkullBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SkullBlockEntity.java
@@ -49,9 +_,9 @@
         }
     };
     @Nullable
-    public ResolvableProfile owner;
+    private ResolvableProfile owner;
     @Nullable
-    public ResourceLocation noteBlockSound;
+    private ResourceLocation noteBlockSound;
     private int animationTickCount;
     private boolean isAnimating;
     @Nullable
