--- a/net/minecraft/world/level/block/entity/TestInstanceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TestInstanceBlockEntity.java
@@ -341,7 +_,7 @@
     }
 
     private void removeEntities() {
-        this.level.getEntities(null, this.getTestBounds()).stream().filter(entity -> !(entity instanceof Player)).forEach(Entity::discard);
+        this.level.getEntities(null, this.getTestBounds()).stream().filter(entity -> !(entity instanceof Player)).forEach(entity -> entity.discard(null)); // CraftBukkit - add Bukkit remove cause
     }
 
     private void forceLoadChunks() {
