--- a/net/minecraft/world/level/block/entity/TestInstanceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TestInstanceBlockEntity.java
@@ -99,7 +_,7 @@
     }
 
     private static Optional<StructureTemplate> getStructureTemplate(ServerLevel level, ResourceKey<GameTestInstance> testKey) {
-        return level.registryAccess().get(testKey).map(test -> test.value().structure()).flatMap(template -> level.getStructureTemplateManager().get(template));
+        return level.registryAccess().get(testKey).map(test -> test.value().structure()).flatMap(template -> level.getStructureManager().get(template));
     }
 
     public Optional<ResourceKey<GameTestInstance>> test() {
@@ -161,6 +_,7 @@
 
     @Override
     protected void loadAdditional(ValueInput input) {
+        super.loadAdditional(input); // Paper - load the PDC
         input.read("data", TestInstanceBlockEntity.Data.CODEC).ifPresent(this::set);
         this.errorMarkers.clear();
         this.errorMarkers.addAll(input.read("errors", TestInstanceBlockEntity.ErrorMarker.LIST_CODEC).orElse(List.of()));
@@ -168,6 +_,7 @@
 
     @Override
     protected void saveAdditional(ValueOutput output) {
+        super.saveAdditional(output); // Paper - save the PDC
         output.store("data", TestInstanceBlockEntity.Data.CODEC, this.data);
         if (!this.errorMarkers.isEmpty()) {
             output.store("errors", TestInstanceBlockEntity.ErrorMarker.LIST_CODEC, this.errorMarkers);
@@ -258,7 +_,7 @@
     }
 
     public static boolean export(ServerLevel level, Identifier structureId, Consumer<Component> feedbackOutput) {
-        StructureTemplateManager structureManager = level.getStructureTemplateManager();
+        StructureTemplateManager structureManager = level.getStructureManager();
         TemplatePathFactory testTemplatePathFactory = structureManager.testTemplates();
         if (testTemplatePathFactory == null) {
             feedbackOutput.accept(Component.literal("Test structure exporting is disabled").withStyle(ChatFormatting.RED));
@@ -342,7 +_,7 @@
     }
 
     private void removeEntities() {
-        this.level.getEntities(null, this.getTestBounds()).stream().filter(entity -> !(entity instanceof Player)).forEach(Entity::discard);
+        this.level.getEntities(null, this.getTestBounds()).stream().filter(entity -> !(entity instanceof Player)).forEach(entity -> entity.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD)); // Paper
     }
 
     private void forceLoadChunks() {
