--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -18,6 +_,7 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.LevelChunk;
+import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.levelgen.feature.EndGatewayFeature;
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
@@ -127,7 +_,7 @@
     }
 
     public @Nullable Vec3 getPortalPosition(ServerLevel currentLevel, BlockPos portalEntryPos) {
-        if (this.exitPortal == null && currentLevel.dimension() == Level.END) {
+        if (this.exitPortal == null && currentLevel.getTypeKey() == LevelStem.END) { // CraftBukkit - work in alternate worlds
             BlockPos exitPortalPos = findOrCreateValidTeleportPos(currentLevel, portalEntryPos);
             exitPortalPos = exitPortalPos.above(10);
             LOGGER.debug("Creating portal at {}", exitPortalPos);
@@ -255,7 +_,7 @@
 
     @Override
     public boolean shouldRenderFace(Direction direction) {
-        return Block.shouldRenderFace(this.getBlockState(), this.level.getBlockState(this.getBlockPos().relative(direction)), direction);
+        return Block.shouldRenderFace(this.level, this.worldPosition, this.getBlockState(), this.level.getBlockState(this.getBlockPos().relative(direction)), direction);
     }
 
     public int getParticleAmount() {
