--- a/net/minecraft/world/level/block/entity/TrialSpawnerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TrialSpawnerBlockEntity.java
@@ -22,7 +_,7 @@
 
 public class TrialSpawnerBlockEntity extends BlockEntity implements Spawner, TrialSpawner.StateAccessor {
     private static final Logger LOGGER = LogUtils.getLogger();
-    public TrialSpawner trialSpawner;
+    private TrialSpawner trialSpawner;
 
     public TrialSpawnerBlockEntity(BlockPos p_312399_, BlockState p_312206_) {
         super(BlockEntityType.TRIAL_SPAWNER, p_312399_, p_312206_);
