--- a/net/minecraft/world/level/block/entity/TrialSpawnerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TrialSpawnerBlockEntity.java
@@ -21,7 +_,7 @@
 import net.minecraft.world.level.storage.ValueOutput;
 
 public class TrialSpawnerBlockEntity extends BlockEntity implements TrialSpawner.StateAccessor, Spawner {
-    private final TrialSpawner trialSpawner = this.createDefaultSpawner();
+    public final TrialSpawner trialSpawner = this.createDefaultSpawner();
 
     public TrialSpawnerBlockEntity(BlockPos worldPosition, BlockState blockState) {
         super(BlockEntityTypes.TRIAL_SPAWNER, worldPosition, blockState);
