--- a/net/minecraft/world/timeline/Timeline.java
+++ b/net/minecraft/world/timeline/Timeline.java
@@ -17,7 +_,7 @@
 import net.minecraft.core.Holder;
 import net.minecraft.core.Registry;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
+import net.minecraft.resources.RegistryFixedCodec;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.KeyframeTrack;
@@ -30,7 +_,7 @@
 import net.minecraft.world.clock.WorldClock;
 
 public class Timeline {
-    public static final Codec<Holder<Timeline>> CODEC = RegistryCodecs.holder(Registries.TIMELINE);
+    public static final Codec<Holder<Timeline>> CODEC = RegistryFixedCodec.create(Registries.TIMELINE);
     private static final Codec<Map<EnvironmentAttribute<?>, AttributeTrack<?, ?>>> TRACKS_CODEC = Codec.dispatchedMap(
         EnvironmentAttributes.CODEC, Util.memoize(AttributeTrack::createCodec)
     );
