--- a/net/minecraft/client/data/models/EquipmentAssetProvider.java
+++ b/net/minecraft/client/data/models/EquipmentAssetProvider.java
@@ -42,24 +_,27 @@
         );
         consumer.accept(EquipmentAssets.CHAINMAIL, onlyHumanoid("chainmail").build());
         consumer.accept(
-            EquipmentAssets.COPPER, humanoidAndMountArmor("copper").replaceTrimPalette(TrimMaterials.COPPER, TrimMaterials.Palette.COPPER_DARKER.id()).build()
-        );
-        consumer.accept(
-            EquipmentAssets.IRON, humanoidAndMountArmor("iron").replaceTrimPalette(TrimMaterials.IRON, TrimMaterials.Palette.IRON_DARKER.id()).build()
-        );
-        consumer.accept(
-            EquipmentAssets.GOLD, humanoidAndMountArmor("gold").replaceTrimPalette(TrimMaterials.GOLD, TrimMaterials.Palette.GOLD_DARKER.id()).build()
+            EquipmentAssets.COPPER,
+            humanoidAndMountArmor("copper").replaceTrimPalette(TrimMaterials.Palette.COPPER.id(), TrimMaterials.Palette.COPPER_DARKER.id()).build()
+        );
+        consumer.accept(
+            EquipmentAssets.IRON,
+            humanoidAndMountArmor("iron").replaceTrimPalette(TrimMaterials.Palette.IRON.id(), TrimMaterials.Palette.IRON_DARKER.id()).build()
+        );
+        consumer.accept(
+            EquipmentAssets.GOLD,
+            humanoidAndMountArmor("gold").replaceTrimPalette(TrimMaterials.Palette.GOLD.id(), TrimMaterials.Palette.GOLD_DARKER.id()).build()
         );
         consumer.accept(
             EquipmentAssets.DIAMOND,
-            humanoidAndMountArmor("diamond").replaceTrimPalette(TrimMaterials.DIAMOND, TrimMaterials.Palette.DIAMOND_DARKER.id()).build()
+            humanoidAndMountArmor("diamond").replaceTrimPalette(TrimMaterials.Palette.DIAMOND.id(), TrimMaterials.Palette.DIAMOND_DARKER.id()).build()
         );
         consumer.accept(
             EquipmentAssets.TURTLE_SCUTE, EquipmentClientInfo.builder().addMainHumanoidLayer(Identifier.withDefaultNamespace("turtle_scute"), false).build()
         );
         consumer.accept(
             EquipmentAssets.NETHERITE,
-            humanoidAndMountArmor("netherite").replaceTrimPalette(TrimMaterials.NETHERITE, TrimMaterials.Palette.NETHERITE_DARKER.id()).build()
+            humanoidAndMountArmor("netherite").replaceTrimPalette(TrimMaterials.Palette.NETHERITE.id(), TrimMaterials.Palette.NETHERITE_DARKER.id()).build()
         );
         consumer.accept(
             EquipmentAssets.ARMADILLO_SCUTE,
