--- a/net/minecraft/client/data/models/ItemModelGenerators.java
+++ b/net/minecraft/client/data/models/ItemModelGenerators.java
@@ -8,6 +_,7 @@
 import net.minecraft.client.color.item.Dye;
 import net.minecraft.client.color.item.Firework;
 import net.minecraft.client.color.item.ItemTintSource;
+import net.minecraft.client.color.item.MapColor;
 import net.minecraft.client.color.item.Potion;
 import net.minecraft.client.data.models.model.ItemModelUtils;
 import net.minecraft.client.data.models.model.ModelInstance;
@@ -107,11 +_,6 @@
         this.itemModelOutput.accept(item, ItemModelUtils.plainModel(this.createFlatItemModel(item, textureDonor, template)));
     }
 
-    private void generateFlatItem(Item item, Identifier textureId, ModelTemplate template) {
-        Identifier model = template.create(ModelLocationUtils.getModelLocation(item), TextureMapping.layer0(new Material(textureId)), this.modelOutput);
-        this.itemModelOutput.accept(item, ItemModelUtils.plainModel(model));
-    }
-
     private void generateItemWithTintedOverlay(Item item, ItemTintSource overlayTint) {
         this.generateItemWithTintedOverlay(item, "_overlay", overlayTint);
     }
@@ -773,23 +_,7 @@
         this.generateFlatItem(Items.OMINOUS_TRIAL_KEY, ModelTemplates.FLAT_ITEM);
         this.generateFlatItem(Items.OMINOUS_BOTTLE, ModelTemplates.FLAT_ITEM);
         this.generateItemWithTintedOverlay(Items.FIREWORK_STAR, new Firework());
-        this.generateFlatItem(Items.FILLED_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.OCEAN_MONUMENT_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.WOODLAND_MANSION_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.BURIED_TRIAL_CHAMBERS_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.JUNGLE_PYRAMID_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.SWAMP_HUT_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.DESERT_VILLAGE_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.PLAINS_VILLAGE_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.SAVANNA_VILLAGE_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.SNOWY_VILLAGE_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.TAIGA_VILLAGE_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.BURIED_TREASURE_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.BURIED_ANCIENT_CITY_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.BURIED_MINESHAFT_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.DESERT_PYRAMID_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.ABANDONED_CAMP_MAP, ModelTemplates.FLAT_ITEM);
-        this.generateFlatItem(Items.WARM_OCEAN_RUINS_MAP, ModelTemplates.FLAT_ITEM);
+        this.generateItemWithTintedOverlay(Items.FILLED_MAP, "_markings", new MapColor());
         this.generateBundleModels(Items.BUNDLE);
         Items.DYED_BUNDLE.forEach(this::generateBundleModels);
         this.generateSpyglass(Items.SPYGLASS);
