--- a/net/minecraft/client/data/models/ModelProvider.java
+++ b/net/minecraft/client/data/models/ModelProvider.java
@@ -29,10 +_,8 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
-import net.neoforged.neoforge.client.extensions.IModelProviderExtension;
-
 @OnlyIn(Dist.CLIENT)
-public class ModelProvider implements DataProvider, IModelProviderExtension {
+public class ModelProvider implements DataProvider {
     private final PackOutput.PathProvider blockStatePathProvider;
     private final PackOutput.PathProvider itemInfoPathProvider;
     private final PackOutput.PathProvider modelPathProvider;
@@ -60,7 +_,7 @@
     }
 
     @Override
-    public String getName() {
+    public final String getName() {
         return "Model Definitions";
     }
 
@@ -102,7 +_,7 @@
             this.register(item, new ClientItem(model, properties));
         }
 
-        public void register(Item item, ClientItem itemInfo) {
+        private void register(Item item, ClientItem itemInfo) {
             ClientItem prev = this.itemInfos.put(item, itemInfo);
             if (prev != null) {
                 throw new IllegalStateException("Duplicate item model definition for " + item);
