--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -566,13 +_,13 @@
 
     @Nullable
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_31636_) {
+    public SoundEvent getHurtSound(DamageSource p_31636_) {
         return SoundEvents.ARMOR_STAND_HIT;
     }
 
     @Nullable
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.ARMOR_STAND_BREAK;
     }
 
