--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -48,6 +_,14 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.craftbukkit.CraftEquipmentSlot;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerArmorStandManipulateEvent;
+// CraftBukkit end
+
 import org.jspecify.annotations.Nullable;
 
 public class ArmorStand extends LivingEntity {
@@ -88,6 +_,12 @@
     private boolean invisible = false;
     public long lastHit;
     public int disabledSlots = 0;
+    public boolean canMove = true; // Paper
+    // Paper start - Allow ArmorStands not to tick
+    public boolean canTick = true;
+    public boolean canTickSetByAPI = false;
+    private boolean noTickEquipmentDirty = false;
+    // Paper end - Allow ArmorStands not to tick
 
     public ArmorStand(EntityType<? extends ArmorStand> type, Level level) {
         super(type, level);
@@ -102,6 +_,13 @@
         return createLivingAttributes().add(Attributes.STEP_HEIGHT, 0.0);
     }
 
+    // CraftBukkit start - SPIGOT-3607, SPIGOT-3637
+    @Override
+    public float getBukkitYaw() {
+        return this.getYRot();
+    }
+    // CraftBukkit end
+
     @Override
     public void refreshDimensions() {
         double oldX = this.getX();
@@ -176,7 +_,7 @@
 
     @Override
     protected void pushEntities() {
-        for (Entity entity : this.level().getEntities(this, this.getBoundingBox(), RIDABLE_MINECARTS)) {
+        for (Entity entity : this.level().getEntitiesOfClass(AbstractMinecart.class, this.getBoundingBox(), RIDABLE_MINECARTS)) { // Paper - optimise collisions
             if (this.distanceToSqr(entity) <= 0.2) {
                 entity.push(this);
             }
@@ -250,57 +_,68 @@
         ItemStack itemStack = this.getItemBySlot(slot);
         if (!itemStack.isEmpty() && (this.disabledSlots & 1 << slot.getFilterBit(8)) != 0) {
             return false;
-        }
-
-        if (itemStack.isEmpty() && (this.disabledSlots & 1 << slot.getFilterBit(16)) != 0) {
+        } else if (itemStack.isEmpty() && (this.disabledSlots & 1 << slot.getFilterBit(16)) != 0) {
             return false;
-        }
+            // CraftBukkit start
+        } else {
+            org.bukkit.inventory.ItemStack armorStandItem = CraftItemStack.asCraftMirror(itemStack);
+            org.bukkit.inventory.ItemStack playerHeldItem = CraftItemStack.asCraftMirror(playerItemStack);
+
+            org.bukkit.entity.Player bukkitPlayer = (org.bukkit.entity.Player) player.getBukkitEntity();
+            org.bukkit.entity.ArmorStand self = (org.bukkit.entity.ArmorStand) this.getBukkitEntity();
+
+            org.bukkit.inventory.EquipmentSlot bukkitSlot = CraftEquipmentSlot.getSlot(slot);
+            org.bukkit.inventory.EquipmentSlot bukkitHand = CraftEquipmentSlot.getHand(hand);
+            PlayerArmorStandManipulateEvent armorStandManipulateEvent = new PlayerArmorStandManipulateEvent(bukkitPlayer, self, playerHeldItem, armorStandItem, bukkitSlot, bukkitHand);
+            this.level().getCraftServer().getPluginManager().callEvent(armorStandManipulateEvent);
+
+            if (armorStandManipulateEvent.isCancelled()) {
+                return true;
+            }
 
         if (player.hasInfiniteMaterials() && itemStack.isEmpty() && !playerItemStack.isEmpty()) {
+            // CraftBukkit end
             this.setItemSlot(slot, playerItemStack.copyWithCount(1));
             return true;
-        }
-
-        if (playerItemStack.isEmpty() || playerItemStack.getCount() <= 1) {
+        } else if (playerItemStack.isEmpty() || playerItemStack.getCount() <= 1) {
             this.setItemSlot(slot, playerItemStack);
             player.setItemInHand(hand, itemStack);
             return true;
-        }
-
-        if (!itemStack.isEmpty()) {
+        } else if (!itemStack.isEmpty()) {
             return false;
+        } else {
+            this.setItemSlot(slot, playerItemStack.split(1));
+            return true;
         }
-
-        this.setItemSlot(slot, playerItemStack.split(1));
-        return true;
+        } // CraftBukkit
     }
 
     @Override
     public boolean hurtServer(ServerLevel level, DamageSource source, float damage) {
         if (this.isRemoved()) {
             return false;
-        }
-
-        if (!level.getGameRules().get(GameRules.MOB_GRIEFING) && source.getEntity() instanceof Mob) {
-            return false;
-        }
-
-        if (source.is(DamageTypeTags.BYPASSES_INVULNERABILITY)) {
-            this.kill(level);
-            return false;
-        }
-
-        if (this.isInvulnerableTo(level, source) || this.invisible || this.isMarker()) {
-            return false;
-        }
-
-        if (source.is(DamageTypeTags.IS_EXPLOSION)) {
+        } else if (!level.getGameRules().get(GameRules.MOB_GRIEFING) && source.getEntity() instanceof Mob) {
+            return false;
+        } else if (source.is(DamageTypeTags.BYPASSES_INVULNERABILITY)) {
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, damage)) {
+                return false;
+            }
+            this.kill(level, source); // CraftBukkit
+            // CraftBukkit end
+            return false;
+        } else if (this.isInvulnerableTo(level, source) || (true || !this.invisible) && !this.isMarker()) { // CraftBukkit
+            return false;
+        } else if (source.is(DamageTypeTags.IS_EXPLOSION)) {
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, damage, true, this.invisible)) {
+                return false;
+            }
+            // CraftBukkit end
             this.brokenByAnything(level, source);
-            this.kill(level);
+            this.kill(level, source); // CraftBukkit
             return false;
-        }
-
-        if (source.is(DamageTypeTags.IGNITES_ARMOR_STANDS)) {
+        } else if (source.is(DamageTypeTags.IGNITES_ARMOR_STANDS)) {
             if (this.isOnFire()) {
                 this.causeDamage(level, source, 0.15F);
             } else {
@@ -334,7 +_,7 @@
                 } else {
                     this.brokenByPlayer(level, source);
                     this.showBreakingParticles();
-                    this.kill(level);
+                    this.discard(EntityRemoveEvent.Cause.DEATH); // CraftBukkit - SPIGOT-4890: remain as this.discard() since above damagesource method will call death event
                 }
 
                 return true;
@@ -386,31 +_,45 @@
         float health = this.getHealth();
         health -= dmg;
         if (health <= 0.5F) {
-            this.brokenByAnything(level, source);
-            this.kill(level);
+            // Paper start - avoid duplicate event call
+            org.bukkit.event.entity.EntityDeathEvent event = this.brokenByAnything(level, source);
+            if (!event.isCancelled()) this.kill(level, source, false); // CraftBukkit
+            // Paper end
         } else {
             this.setHealth(health);
             this.gameEvent(GameEvent.ENTITY_DAMAGE, source.getEntity());
         }
     }
 
-    private void brokenByPlayer(ServerLevel level, DamageSource source) {
+    private org.bukkit.event.entity.EntityDeathEvent brokenByPlayer(ServerLevel level, DamageSource source) { // Paper
         ItemStack result = new ItemStack(Items.ARMOR_STAND);
         result.set(DataComponents.CUSTOM_NAME, this.getCustomName());
-        Block.popResource(this.level(), this.blockPosition(), result);
-        this.brokenByAnything(level, source);
+        this.deathDropItems = new java.util.ArrayList<>(); // Paper
+        this.deathDropItems.add(new DefaultDrop(result, stack -> Block.popResource(this.level(), this.blockPosition(), stack))); // CraftBukkit - add to drops // Paper - Restore vanilla drops behavior
+        return this.brokenByAnything(level, source); // Paper
     }
 
-    private void brokenByAnything(ServerLevel level, DamageSource source) {
+    private org.bukkit.event.entity.EntityDeathEvent brokenByAnything(ServerLevel level, DamageSource source) { // Paper
         this.playBrokenSound();
+        if (this.deathDropItems == null) this.deathDropItems = new java.util.ArrayList<>(); // Paper
         this.dropAllDeathLoot(level, source);
 
         for (EquipmentSlot slot : EquipmentSlot.VALUES) {
-            ItemStack itemStack = this.equipment.set(slot, ItemStack.EMPTY);
+            this.postDeathEventTasks.add(() -> this.equipment.set(slot, ItemStack.EMPTY)); // Paper - move equipment removal past event call
+            ItemStack itemStack = this.equipment.get(slot); // Paper
             if (!itemStack.isEmpty() && !EnchantmentHelper.has(itemStack, EnchantmentEffectComponents.PREVENT_EQUIPMENT_DROP)) {
-                Block.popResource(this.level(), this.blockPosition().above(), itemStack);
+                this.deathDropItems.add(new DefaultDrop(itemStack, stack -> Block.popResource(this.level(), this.blockPosition().above(), stack))); // CraftBukkit - add to drops // Paper - Restore vanilla drops behavior; mirror so we can destroy it later - though this call site was safe & spawn drops correctly}
             }
         }
+        // Paper start - call EntityDeathEvent
+        org.bukkit.event.entity.EntityDeathEvent deathEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(level, this, source, this.deathDropItems, true);
+        this.deathDropItems = null;
+        if (!deathEvent.isCancelled()) {
+            this.postDeathEventTasks.forEach(Runnable::run);
+        }
+        this.postDeathEventTasks.clear();
+        return deathEvent;
+        // Paper end - call EntityDeathEvent
     }
 
     private void playBrokenSound() {
@@ -458,9 +_,29 @@
         return this.isSmall();
     }
 
+    // CraftBukkit start
+    @Override
+    public boolean shouldDropExperience() {
+        return true; // MC-157395, SPIGOT-5193 even baby (small) armor stands should drop
+    }
+    // CraftBukkit end
+
     @Override
     public void kill(ServerLevel level) {
-        this.remove(Entity.RemovalReason.KILLED);
+        // Paper start - pass damage source for event
+        this.kill(level, null);
+    }
+
+    public void kill(final ServerLevel level, final @Nullable DamageSource source) {
+        this.kill(level, source, true);
+    }
+
+    public void kill(final ServerLevel level, final @Nullable DamageSource source, final boolean callEvent) {
+        if (callEvent && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityDeathEvent(level, this, (source == null ? this.damageSources().genericKill() : source))) {
+            return;
+        }
+        // Paper end
+        this.remove(Entity.RemovalReason.KILLED, EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause
         this.gameEvent(GameEvent.ENTITY_DIE);
     }
 
