--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.decoration;
 
+import java.util.Collection;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -30,6 +_,7 @@
 import net.minecraft.world.entity.Pose;
 import net.minecraft.world.entity.ai.attributes.AttributeSupplier;
 import net.minecraft.world.entity.ai.attributes.Attributes;
+import net.minecraft.world.entity.item.ItemEntity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.vehicle.AbstractMinecart;
 import net.minecraft.world.item.ItemStack;
@@ -43,6 +_,9 @@
 import net.minecraft.world.level.material.PushReaction;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.CraftEquipmentSlot;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.craftbukkit.inventory.CraftItemStack;
 
 public class ArmorStand extends LivingEntity {
     public static final int WOBBLE_TIME = 5;
@@ -73,7 +_,7 @@
     public static final EntityDataAccessor<Rotations> DATA_LEFT_LEG_POSE = SynchedEntityData.defineId(ArmorStand.class, EntityDataSerializers.ROTATIONS);
     public static final EntityDataAccessor<Rotations> DATA_RIGHT_LEG_POSE = SynchedEntityData.defineId(ArmorStand.class, EntityDataSerializers.ROTATIONS);
     private static final Predicate<Entity> RIDABLE_MINECARTS = p_31582_ -> p_31582_ instanceof AbstractMinecart
-            && ((AbstractMinecart)p_31582_).getMinecartType() == AbstractMinecart.Type.RIDEABLE;
+            && ((AbstractMinecart)p_31582_).canBeRidden();
     private final NonNullList<ItemStack> handItems = NonNullList.withSize(2, ItemStack.EMPTY);
     private final NonNullList<ItemStack> armorItems = NonNullList.withSize(4, ItemStack.EMPTY);
     private boolean invisible;
@@ -85,9 +_,17 @@
     public Rotations rightArmPose = DEFAULT_RIGHT_ARM_POSE;
     public Rotations leftLegPose = DEFAULT_LEFT_LEG_POSE;
     public Rotations rightLegPose = DEFAULT_RIGHT_LEG_POSE;
+    public boolean canMove = true; // Paper
+    // Paper start - Allow ArmorStands not to tick
+    public boolean canTick = true;
+    public boolean canTickSetByAPI = false;
+    private boolean noTickPoseDirty = false;
+    private boolean noTickEquipmentDirty = false;
+    // Paper end - Allow ArmorStands not to tick
 
     public ArmorStand(EntityType<? extends ArmorStand> p_31553_, Level p_31554_) {
         super(p_31553_, p_31554_);
+        if (p_31554_ != null) this.canTick = p_31554_.paperConfig().entities.armorStands.tick; // Paper - Allow ArmorStands not to tick
     }
 
     public ArmorStand(Level p_31556_, double p_31557_, double p_31558_, double p_31559_) {
@@ -99,6 +_,13 @@
         return createLivingAttributes().add(Attributes.STEP_HEIGHT, 0.0);
     }
 
+    // CraftBukkit start - SPIGOT-3607, SPIGOT-3637
+    @Override
+    public float getBukkitYaw() {
+        return this.getYRot();
+    }
+    // CraftBukkit end
+
     @Override
     public void refreshDimensions() {
         double d0 = this.getX();
@@ -158,14 +_,22 @@
 
     @Override
     public void setItemSlot(EquipmentSlot p_31584_, ItemStack p_31585_) {
+        // CraftBukkit start
+        this.setItemSlot(p_31584_, p_31585_, false);
+    }
+
+    @Override
+    public void setItemSlot(EquipmentSlot p_31584_, ItemStack p_31585_, boolean silent) {
         this.verifyEquippedItem(p_31585_);
         switch (p_31584_.getType()) {
             case HAND:
-                this.onEquipItem(p_31584_, this.handItems.set(p_31584_.getIndex(), p_31585_), p_31585_);
+                this.onEquipItem(p_31584_, this.handItems.set(p_31584_.getIndex(), p_31585_), p_31585_, silent);
                 break;
             case HUMANOID_ARMOR:
-                this.onEquipItem(p_31584_, this.armorItems.set(p_31584_.getIndex(), p_31585_), p_31585_);
+                this.onEquipItem(p_31584_, this.armorItems.set(p_31584_.getIndex(), p_31585_), p_31585_, silent);
         }
+
+        this.noTickEquipmentDirty = true; // Paper - Allow ArmorStands not to tick; Still update equipment
     }
 
     @Override
@@ -201,6 +_,7 @@
         }
 
         p_31619_.put("Pose", this.writePose());
+        if (this.canTickSetByAPI) p_31619_.putBoolean("Paper.CanTickOverride", this.canTick); // Paper - Allow ArmorStands not to tick
     }
 
     @Override
@@ -231,6 +_,12 @@
         this.setNoBasePlate(p_31600_.getBoolean("NoBasePlate"));
         this.setMarker(p_31600_.getBoolean("Marker"));
         this.noPhysics = !this.hasPhysics();
+        // Paper start - Allow ArmorStands not to tick
+        if (p_31600_.contains("Paper.CanTickOverride")) {
+            this.canTick = p_31600_.getBoolean("Paper.CanTickOverride");
+            this.canTickSetByAPI = true;
+        }
+        // Paper end - Allow ArmorStands not to tick
         CompoundTag compoundtag1 = p_31600_.getCompound("Pose");
         this.readPose(compoundtag1);
     }
@@ -290,6 +_,7 @@
 
     @Override
     protected void pushEntities() {
+        if (!this.level().paperConfig().entities.armorStands.doCollisionEntityLookups) return; // Paper - Option to prevent armor stands from doing entity lookups
         for (Entity entity : this.level().getEntities(this, this.getBoundingBox(), RIDABLE_MINECARTS)) {
             if (this.distanceToSqr(entity) <= 0.2) {
                 entity.push(this);
@@ -352,7 +_,7 @@
         return equipmentslot;
     }
 
-    private boolean isDisabled(EquipmentSlot p_31627_) {
+    public boolean isDisabled(EquipmentSlot p_31627_) {
         return (this.disabledSlots & 1 << p_31627_.getFilterFlag()) != 0 || p_31627_.getType() == EquipmentSlot.Type.HAND && !this.isShowArms();
     }
 
@@ -362,7 +_,25 @@
             return false;
         } else if (itemstack.isEmpty() && (this.disabledSlots & 1 << p_31590_.getFilterFlag() + 16) != 0) {
             return false;
-        } else if (p_31589_.hasInfiniteMaterials() && itemstack.isEmpty() && !p_31591_.isEmpty()) {
+            // CraftBukkit start
+        } else {
+            org.bukkit.inventory.ItemStack armorStandItem = CraftItemStack.asCraftMirror(itemstack);
+            org.bukkit.inventory.ItemStack playerHeldItem = CraftItemStack.asCraftMirror(p_31591_);
+
+            org.bukkit.entity.Player player = (org.bukkit.entity.Player) p_31589_.getBukkitEntity();
+            org.bukkit.entity.ArmorStand self = (org.bukkit.entity.ArmorStand) this.getBukkitEntity();
+
+            org.bukkit.inventory.EquipmentSlot slot = CraftEquipmentSlot.getSlot(p_31590_);
+            org.bukkit.inventory.EquipmentSlot hand = CraftEquipmentSlot.getHand(p_31592_);
+            org.bukkit.event.player.PlayerArmorStandManipulateEvent armorStandManipulateEvent = new org.bukkit.event.player.PlayerArmorStandManipulateEvent(player, self, playerHeldItem, armorStandItem, slot, hand);
+            this.level().getCraftServer().getPluginManager().callEvent(armorStandManipulateEvent);
+
+            if (armorStandManipulateEvent.isCancelled()) {
+                return true;
+            }
+
+            if (p_31589_.hasInfiniteMaterials() && itemstack.isEmpty() && !p_31591_.isEmpty()) {
+                // CraftBukkit end
             this.setItemSlot(p_31590_, p_31591_.copyWithCount(1));
             return true;
         } else if (p_31591_.isEmpty() || p_31591_.getCount() <= 1) {
@@ -375,6 +_,7 @@
             this.setItemSlot(p_31590_, p_31591_.split(1));
             return true;
         }
+        } // CraftBukkit
     }
 
     @Override
@@ -383,13 +_,23 @@
             return false;
         } else if (this.level() instanceof ServerLevel serverlevel) {
             if (p_31579_.is(DamageTypeTags.BYPASSES_INVULNERABILITY)) {
-                this.kill();
+                // CraftBukkit start
+                if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31579_, p_31580_)) {
+                    return false;
+                }
+                this.kill(p_31579_); // CraftBukkit
+                // CraftBukkit end
                 return false;
-            } else if (this.isInvulnerableTo(p_31579_) || this.invisible || this.isMarker()) {
+             } else if (this.isInvulnerableTo(p_31579_) || this.invisible || this.isMarker()) {
+                // CraftBukkit start
+                if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31579_, p_31580_, true, this.invisible)) {
+                    return false;
+                }
+                // CraftBukkit end
                 return false;
             } else if (p_31579_.is(DamageTypeTags.IS_EXPLOSION)) {
                 this.brokenByAnything(serverlevel, p_31579_);
-                this.kill();
+                this.kill(p_31579_); // CraftBukkit
                 return false;
             } else if (p_31579_.is(DamageTypeTags.IGNITES_ARMOR_STANDS)) {
                 if (this.isOnFire()) {
@@ -415,7 +_,7 @@
                     if (p_31579_.isCreativePlayer()) {
                         this.playBrokenSound();
                         this.showBreakingParticles();
-                        this.kill();
+                        this.kill(p_31579_); // CraftBukkit
                         return true;
                     } else {
                         long i = serverlevel.getGameTime();
@@ -426,7 +_,7 @@
                         } else {
                             this.brokenByPlayer(serverlevel, p_31579_);
                             this.showBreakingParticles();
-                            this.kill();
+                            this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DEATH); // CraftBukkit - SPIGOT-4890: remain as this.discard() since above damagesource method will call death event
                         }
 
                         return true;
@@ -483,7 +_,7 @@
         f -= p_31650_;
         if (f <= 0.5F) {
             this.brokenByAnything(p_348633_, p_31649_);
-            this.kill();
+            this.kill(p_31649_); // CraftBukkit
         } else {
             this.setHealth(f);
             this.gameEvent(GameEvent.ENTITY_DAMAGE, p_31649_.getEntity());
@@ -550,7 +_,29 @@
 
     @Override
     public void tick() {
+        // Paper start - Allow ArmorStands not to tick
+        if (!this.canTick) {
+            if (this.noTickPoseDirty) {
+                this.noTickPoseDirty = false;
+                this.updatePose();
+            }
+
+            if (this.noTickEquipmentDirty) {
+                this.noTickEquipmentDirty = false;
+                this.detectEquipmentUpdatesPublic();
+            }
+
+            return;
+        }
+        // Paper end - Allow ArmorStands not to tick
+
         super.tick();
+        // Paper start - Allow ArmorStands not to tick
+        updatePose();
+    }
+
+    public void updatePose() {
+        // Paper end - Allow ArmorStands not to tick
         Rotations rotations = this.entityData.get(DATA_HEAD_POSE);
         if (!this.headPose.equals(rotations)) {
             this.setHeadPose(rotations);
@@ -598,9 +_,29 @@
         return this.isSmall();
     }
 
+    // CraftBukkit start
+    @Override
+    public boolean shouldDropExperience() {
+        return true; // MC-157395, SPIGOT-5193 even baby (small) armor stands should drop
+    }
+    // CraftBukkit end
+
     @Override
     public void kill() {
-        this.remove(Entity.RemovalReason.KILLED);
+        Collection<ItemEntity> captureDrops = this.captureDrops(null);
+        CraftEventFactory.callEntityDeathEvent(this, this.damageSources().genericKill(), captureDrops, com.google.common.util.concurrent.Runnables.doNothing()); // CraftBukkit - call event
+        setRemoved$cause(org.bukkit.event.entity.EntityRemoveEvent.Cause.DEATH);
+        this.remove(Entity.RemovalReason.KILLED); // CraftBukkit - add Bukkit remove cause
+        // CraftBukkit end
+        this.gameEvent(GameEvent.ENTITY_DIE);
+    }
+
+    public void kill(DamageSource damageSource) {
+        Collection<ItemEntity> captureDrops = this.captureDrops(null);
+        CraftEventFactory.callEntityDeathEvent(this, (damageSource == null ? this.damageSources().genericKill() : damageSource), captureDrops, com.google.common.util.concurrent.Runnables.doNothing()); // CraftBukkit - call event
+        this.setRemoved$cause(org.bukkit.event.entity.EntityRemoveEvent.Cause.DEATH);
+        this.remove(Entity.RemovalReason.KILLED); // CraftBukkit - add Bukkit remove cause
+        // CraftBukkit end
         this.gameEvent(GameEvent.ENTITY_DIE);
     }
 
@@ -664,31 +_,37 @@
     public void setHeadPose(Rotations p_31598_) {
         this.headPose = p_31598_;
         this.entityData.set(DATA_HEAD_POSE, p_31598_);
+        this.noTickPoseDirty = true; // Paper - Allow updates when not ticking
     }
 
     public void setBodyPose(Rotations p_31617_) {
         this.bodyPose = p_31617_;
         this.entityData.set(DATA_BODY_POSE, p_31617_);
+        this.noTickPoseDirty = true; // Paper - Allow updates when not ticking
     }
 
     public void setLeftArmPose(Rotations p_31624_) {
         this.leftArmPose = p_31624_;
         this.entityData.set(DATA_LEFT_ARM_POSE, p_31624_);
+        this.noTickPoseDirty = true; // Paper - Allow updates when not ticking
     }
 
     public void setRightArmPose(Rotations p_31629_) {
         this.rightArmPose = p_31629_;
         this.entityData.set(DATA_RIGHT_ARM_POSE, p_31629_);
+        this.noTickPoseDirty = true; // Paper - Allow updates when not ticking
     }
 
     public void setLeftLegPose(Rotations p_31640_) {
         this.leftLegPose = p_31640_;
         this.entityData.set(DATA_LEFT_LEG_POSE, p_31640_);
+        this.noTickPoseDirty = true; // Paper - Allow updates when not ticking
     }
 
     public void setRightLegPose(Rotations p_31652_) {
         this.rightLegPose = p_31652_;
         this.entityData.set(DATA_RIGHT_LEG_POSE, p_31652_);
+        this.noTickPoseDirty = true; // Paper - Allow updates when not ticking
     }
 
     public Rotations getHeadPose() {
@@ -743,7 +_,7 @@
 
     @Nullable
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.ARMOR_STAND_BREAK;
     }
 
@@ -820,4 +_,13 @@
     public boolean canBeSeenByAnyone() {
         return !this.isInvisible() && !this.isMarker();
     }
+
+    // Paper start
+    @Override
+    public void move(net.minecraft.world.entity.MoverType type, Vec3 movement) {
+        if (this.canMove) {
+            super.move(type, movement);
+        }
+    }
+    // Paper end
 }
