--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.decoration;
 
+import io.papermc.paper.event.player.PlayerItemFrameChangeEvent;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -38,6 +_,7 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang3.Validate;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class ItemFrame extends HangingEntity {
     public static final EntityDataAccessor<ItemStack> DATA_ITEM = SynchedEntityData.defineId(ItemFrame.class, EntityDataSerializers.ITEM_STACK);
@@ -85,6 +_,15 @@
         this.recalculateBoundingBox();
     }
 
+    public static AABB calculateBoundingBoxStatic(BlockPos blockposition, Direction enumdirection) {
+        Vec3 vec3 = Vec3.atCenterOf(blockposition).relative(enumdirection, -0.46875);
+        Direction.Axis direction$axis = enumdirection.getAxis();
+        double d0 = direction$axis == Direction.Axis.X ? 0.0625 : 0.75;
+        double d1 = direction$axis == Direction.Axis.Y ? 0.0625 : 0.75;
+        double d2 = direction$axis == Direction.Axis.Z ? 0.0625 : 0.75;
+        return AABB.ofSize(vec3, d0, d1, d2);
+    }
+
     @Override
     protected AABB calculateBoundingBox(BlockPos p_345363_, Direction p_344892_) {
         float f = 0.46875F;
@@ -118,7 +_,7 @@
     }
 
     @Override
-    public void push(double p_31817_, double p_31818_, double p_31819_) {
+    public void push(double p_31817_, double p_31818_, double p_31819_, @org.jetbrains.annotations.Nullable Entity pushingEntity) { // Paper - add push source entity param
         if (!this.fixed) {
             super.push(p_31817_, p_31818_, p_31819_);
         }
@@ -138,6 +_,18 @@
             return false;
         } else if (!p_31776_.is(DamageTypeTags.IS_EXPLOSION) && !this.getItem().isEmpty()) {
             if (!this.level().isClientSide) {
+                // CraftBukkit start - fire EntityDamageEvent
+                if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31776_, p_31777_, false) || this.isRemoved()) {
+                    return true;
+                }
+                // CraftBukkit end
+                // Paper start - Add PlayerItemFrameChangeEvent
+                if (p_31776_.getEntity() instanceof Player player) {
+                    var event = new PlayerItemFrameChangeEvent((org.bukkit.entity.Player) player.getBukkitEntity(), (org.bukkit.entity.ItemFrame) this.getBukkitEntity(), this.getItem().asBukkitCopy(), PlayerItemFrameChangeEvent.ItemFrameChangeAction.REMOVE);
+                    if (!event.callEvent()) return true; // return true here because you aren't cancelling the damage, just the change
+                    this.setItem(ItemStack.fromBukkitCopy(event.getItemStack()), false);
+                }
+                // Paper end - Add PlayerItemFrameChangeEvent
                 this.dropItem(p_31776_.getEntity(), false);
                 this.gameEvent(GameEvent.BLOCK_CHANGE, p_31776_.getEntity());
                 this.playSound(this.getRemoveItemSound(), 1.0F, 1.0F);
@@ -212,7 +_,7 @@
     private void removeFramedMap(ItemStack p_31811_) {
         MapId mapid = this.getFramedMapId(p_31811_);
         if (mapid != null) {
-            MapItemSavedData mapitemsaveddata = MapItem.getSavedData(mapid, this.level());
+            MapItemSavedData mapitemsaveddata = MapItem.getSavedData(p_31811_, this.level());
             if (mapitemsaveddata != null) {
                 mapitemsaveddata.removedFromFrame(this.pos, this.getId());
                 mapitemsaveddata.setDirty(true);
@@ -226,6 +_,14 @@
         return this.getEntityData().get(DATA_ITEM);
     }
 
+    // Paper start - Fix MC-123848 (spawn item frame drops above block)
+    @Nullable
+    @Override
+    public net.minecraft.world.entity.item.ItemEntity spawnAtLocation(ItemStack stack) {
+        return this.spawnAtLocation(stack, getDirection().equals(Direction.DOWN) ? -0.6F : 0.0F);
+    }
+    // Paper end
+
     @Nullable
     public MapId getFramedMapId(ItemStack p_353029_) {
         return p_353029_.get(DataComponents.MAP_ID);
@@ -255,6 +_,22 @@
         }
     }
 
+    public void setItem(ItemStack p_31790_, boolean p_31791_, boolean playSound) {
+        if (!p_31790_.isEmpty()) {
+            p_31790_ = p_31790_.copyWithCount(1);
+        }
+
+        this.onItemChanged(p_31790_);
+        this.getEntityData().set(DATA_ITEM, p_31790_);
+        if (!p_31790_.isEmpty() && p_31791_ && playSound) { // CraftBukkit
+            this.playSound(this.getAddItemSound(), 1.0F, 1.0F);
+        }
+
+        if (p_31791_ && this.pos != null) {
+            this.level().updateNeighbourForOutputSignal(this.pos, Blocks.AIR);
+        }
+    }
+
     public SoundEvent getAddItemSound() {
         return SoundEvents.ITEM_FRAME_ADD_ITEM;
     }
@@ -354,11 +_,24 @@
                         }
                     }
 
-                    this.setItem(itemstack);
+                    // Paper start - Add PlayerItemFrameChangeEvent
+                    PlayerItemFrameChangeEvent event = new PlayerItemFrameChangeEvent((org.bukkit.entity.Player) p_31787_.getBukkitEntity(), (org.bukkit.entity.ItemFrame) this.getBukkitEntity(), itemstack.asBukkitCopy(), PlayerItemFrameChangeEvent.ItemFrameChangeAction.PLACE);
+                    if (!event.callEvent()) {
+                        return InteractionResult.FAIL;
+                    }
+                    this.setItem(ItemStack.fromBukkitCopy(event.getItemStack()));
+                    // Paper end - Add PlayerItemFrameChangeEvent
                     this.gameEvent(GameEvent.BLOCK_CHANGE, p_31787_);
                     itemstack.consume(1, p_31787_);
                 }
             } else {
+                // Paper start - Add PlayerItemFrameChangeEvent
+                PlayerItemFrameChangeEvent event = new PlayerItemFrameChangeEvent((org.bukkit.entity.Player) p_31787_.getBukkitEntity(), (org.bukkit.entity.ItemFrame) this.getBukkitEntity(), this.getItem().asBukkitCopy(), PlayerItemFrameChangeEvent.ItemFrameChangeAction.ROTATE);
+                if (!event.callEvent()) {
+                    return InteractionResult.FAIL;
+                }
+                setItem(ItemStack.fromBukkitCopy(event.getItemStack()), false, false);
+                // Paper end - Add PlayerItemFrameChangeEvent
                 this.playSound(this.getRotateItemSound(), 1.0F, 1.0F);
                 this.setRotation(this.getRotation() + 1);
                 this.gameEvent(GameEvent.BLOCK_CHANGE, p_31787_);
