--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.decoration;
 
 import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.component.DataComponents;
@@ -110,7 +_,7 @@
         return this.createBoundingBox(this.pos, this.getDirection(), false);
     }
 
-    private AABB createBoundingBox(BlockPos blockPos, Direction direction, boolean hasFramedMap) {
+    public static AABB createBoundingBox(BlockPos blockPos, Direction direction, boolean hasFramedMap) {
         float shiftToBlockWall = 0.46875F;
         Vec3 position = Vec3.atCenterOf(blockPos).relative(direction, -0.46875);
         float width = hasFramedMap ? 1.0F : 0.75F;
@@ -173,6 +_,11 @@
             if (this.isInvulnerableToBase(source)) {
                 return false;
             } else if (this.shouldDamageDropItem(source)) {
+                // CraftBukkit start - fire EntityDamageEvent
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, damage, false) || this.isRemoved()) {
+                    return true;
+                }
+                // CraftBukkit end
                 this.dropItem(level, source.getEntity(), false);
                 this.gameEvent(GameEvent.BLOCK_CHANGE, source.getEntity());
                 this.playSound(this.getRemoveItemSound(), 1.0F, 1.0F);
@@ -245,7 +_,7 @@
     private void removeFramedMap(ItemStack itemStack) {
         MapId mapId = this.getFramedMapId(itemStack);
         if (mapId != null) {
-            MapItemSavedData mapItemSavedData = MapItem.getSavedData(mapId, this.level());
+            MapItemSavedData mapItemSavedData = MapItem.getSavedData(itemStack, this.level());
             if (mapItemSavedData != null) {
                 mapItemSavedData.removedFromFrame(this.pos, this.getId());
             }
@@ -269,19 +_,31 @@
     }
 
     public void setItem(ItemStack itemStack, boolean updateNeighbours) {
+        // CraftBukkit start
+        this.setItem(itemStack, updateNeighbours, true);
+    }
+
+    public void setItem(ItemStack itemStack, boolean updateNeighbours, boolean playSound) {
+        // CraftBukkit end
         if (!itemStack.isEmpty()) {
             itemStack = itemStack.copyWithCount(1);
         }
 
         this.onItemChanged(itemStack);
         this.getEntityData().set(DATA_ITEM, itemStack);
-        if (!itemStack.isEmpty()) {
+        if (!itemStack.isEmpty() && playSound) { // CraftBukkit
             this.playSound(this.getAddItemSound(), 1.0F, 1.0F);
         }
 
         if (updateNeighbours && this.pos != null) {
             this.level().updateNeighbourForOutputSignal(this.pos, Blocks.AIR);
         }
+    }
+
+    AtomicBoolean setItem$playSound = new AtomicBoolean(true);
+    public void setItem(ItemStack p_31790_, boolean p_31791_, boolean playSound) {
+        setItem$playSound.set(playSound);
+        setItem(p_31790_, p_31791_);
     }
 
     public SoundEvent getAddItemSound() {
