--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -25,6 +_,14 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+
+// CraftBukkit start
+import net.minecraft.network.protocol.game.ClientboundSetEntityLinkPacket;
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 import org.jspecify.annotations.Nullable;
 
 public class LeashFenceKnotEntity extends BlockAttachedEntity {
@@ -76,8 +_,7 @@
         if (this.level().isClientSide()) {
             return InteractionResult.SUCCESS;
         }
-
-        if (player.getItemInHand(hand).is(Items.SHEARS)) {
+        if (player.getItemInHand(hand).canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SHEARS_HARVEST)) {
             InteractionResult result = super.interact(player, hand, location);
             if (result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
                 return result;
@@ -87,7 +_,7 @@
         boolean attachedMob = false;
 
         for (Leashable leashable : Leashable.leashableLeashedTo(player)) {
-            if (leashable.canHaveALeashAttachedTo(this)) {
+            if (leashable.canHaveALeashAttachedTo(this) && org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerLeashEntityEvent(leashable, this, player, hand)) { // Paper - leash event
                 leashable.setLeashedTo(this, true);
                 attachedMob = true;
             }
@@ -96,7 +_,7 @@
         boolean anyDropped = false;
         if (!attachedMob && !player.isSecondaryUseActive()) {
             for (Leashable mob : Leashable.leashableLeashedTo(this)) {
-                if (mob.canHaveALeashAttachedTo(player)) {
+                if (mob.canHaveALeashAttachedTo(player) && org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerLeashEntityEvent(mob, player, player, hand)) { // Paper - leash event
                     mob.setLeashedTo(player, true);
                     anyDropped = true;
                 }
@@ -106,7 +_,6 @@
         if (!attachedMob && !anyDropped) {
             return super.interact(player, hand, location);
         }
-
         this.gameEvent(GameEvent.BLOCK_ATTACH, player);
         this.playSound(SoundEvents.LEAD_TIED);
         return InteractionResult.SUCCESS;
@@ -115,7 +_,7 @@
     @Override
     public void notifyLeasheeRemoved(Leashable entity) {
         if (Leashable.leashableLeashedTo(this).isEmpty()) {
-            this.discard();
+            this.discard(EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
         }
     }
 
