--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -25,6 +_,14 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+
+// CraftBukkit start
+import net.minecraft.network.protocol.game.ClientboundSetEntityLinkPacket;
+import net.minecraft.server.level.ServerPlayer;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 import org.jspecify.annotations.Nullable;
 
 public class LeashFenceKnotEntity extends BlockAttachedEntity {
@@ -75,47 +_,68 @@
     public InteractionResult interact(Player player, InteractionHand hand, Vec3 location) {
         if (this.level().isClientSide()) {
             return InteractionResult.SUCCESS;
-        }
-
-        if (player.getItemInHand(hand).is(Items.SHEARS)) {
-            InteractionResult result = super.interact(player, hand, location);
-            if (result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
-                return result;
-            }
-        }
-
-        boolean attachedMob = false;
-
-        for (Leashable leashable : Leashable.leashableLeashedTo(player)) {
-            if (leashable.canHaveALeashAttachedTo(this)) {
-                leashable.setLeashedTo(this, true);
-                attachedMob = true;
-            }
-        }
-
-        boolean anyDropped = false;
-        if (!attachedMob && !player.isSecondaryUseActive()) {
-            for (Leashable mob : Leashable.leashableLeashedTo(this)) {
-                if (mob.canHaveALeashAttachedTo(player)) {
-                    mob.setLeashedTo(player, true);
-                    anyDropped = true;
-                }
-            }
-        }
-
-        if (!attachedMob && !anyDropped) {
-            return super.interact(player, hand, location);
-        }
-
-        this.gameEvent(GameEvent.BLOCK_ATTACH, player);
-        this.playSound(SoundEvents.LEAD_TIED);
-        return InteractionResult.SUCCESS;
+        } else {
+            if (player.getItemInHand(hand).canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SHEARS_HARVEST)) {
+                InteractionResult result = super.interact(player, hand, location);
+                if (result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
+                    return result;
+                }
+            }
+
+            boolean attachedMob = false;
+
+            for (Leashable leashable : Leashable.leashableLeashedTo(player)) {
+                if (leashable.canHaveALeashAttachedTo(this)) {
+                    // CraftBukkit start
+                    if (leashable instanceof Entity leashed) {
+                        if (CraftEventFactory.callPlayerLeashEntityEvent(leashed, this, player, hand).isCancelled()) {
+                            ((ServerPlayer) player).connection.send(new ClientboundSetEntityLinkPacket(leashed, leashable.getLeashHolder()));
+                            attachedMob = true; // Also set true when the event is cancelled otherwise it tries to unleash the entities
+                            continue;
+                        }
+                    }
+                    // CraftBukkit end
+                    leashable.setLeashedTo(this, true);
+                    attachedMob = true;
+                }
+            }
+
+            boolean anyDropped = false;
+            if (!attachedMob && !player.isSecondaryUseActive()) {
+                for (Leashable leashable1 : Leashable.leashableLeashedTo(this)) {
+                    if (leashable1.canHaveALeashAttachedTo(player)) {
+                        // CraftBukkit start
+                        if (leashable1 instanceof Entity leashed) {
+                            if (CraftEventFactory.callPlayerUnleashEntityEvent(leashed, player, hand).isCancelled()) {
+                                continue;
+                            }
+                            if (CraftEventFactory.callPlayerLeashEntityEvent(leashed, this, player, hand).isCancelled()) {
+                                ((ServerPlayer) player).connection.send(new ClientboundSetEntityLinkPacket(leashed, leashable1.getLeashHolder()));
+                                anyDropped = true; // Also set true when the event is cancelled otherwise it tries to unleash the entities
+                                continue;
+                            }
+                        }
+                        // CraftBukkit end
+                        leashable1.setLeashedTo(player, true);
+                        anyDropped = true;
+                    }
+                }
+            }
+
+            if (!attachedMob && !anyDropped) {
+                return super.interact(player, hand, location);
+            } else {
+                this.gameEvent(GameEvent.BLOCK_ATTACH, player);
+                this.playSound(SoundEvents.LEAD_TIED);
+                return InteractionResult.SUCCESS;
+            }
+        }
     }
 
     @Override
     public void notifyLeasheeRemoved(Leashable entity) {
         if (Leashable.leashableLeashedTo(this).isEmpty()) {
-            this.discard();
+            this.discard(EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
         }
     }
 
