--- a/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
+++ b/net/minecraft/world/entity/decoration/LeashFenceKnotEntity.java
@@ -25,6 +_,8 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityRemoveEvent;
+
 import org.jspecify.annotations.Nullable;
 
 public class LeashFenceKnotEntity extends BlockAttachedEntity {
@@ -77,7 +_,7 @@
             return InteractionResult.SUCCESS;
         }
 
-        if (player.getItemInHand(hand).is(Items.SHEARS)) {
+        if (player.getItemInHand(hand).canPerformAction(net.neoforged.neoforge.common.ItemAbilities.SHEARS_HARVEST)) {
             InteractionResult result = super.interact(player, hand, location);
             if (result instanceof InteractionResult.Success success && success.wasItemInteraction()) {
                 return result;
@@ -87,7 +_,7 @@
         boolean attachedMob = false;
 
         for (Leashable leashable : Leashable.leashableLeashedTo(player)) {
-            if (leashable.canHaveALeashAttachedTo(this)) {
+            if (leashable.canHaveALeashAttachedTo(this) && org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerLeashEntityEvent(leashable, this, player, hand)) { // Paper - leash event
                 leashable.setLeashedTo(this, true);
                 attachedMob = true;
             }
@@ -96,7 +_,7 @@
         boolean anyDropped = false;
         if (!attachedMob && !player.isSecondaryUseActive()) {
             for (Leashable mob : Leashable.leashableLeashedTo(this)) {
-                if (mob.canHaveALeashAttachedTo(player)) {
+                if (mob.canHaveALeashAttachedTo(player) && org.bukkit.craftbukkit.event.CraftEventFactory.handlePlayerLeashEntityEvent(mob, player, player, hand)) { // Paper - leash event
                     mob.setLeashedTo(player, true);
                     anyDropped = true;
                 }
@@ -115,7 +_,7 @@
     @Override
     public void notifyLeasheeRemoved(Leashable entity) {
         if (Leashable.leashableLeashedTo(this).isEmpty()) {
-            this.discard();
+            this.discard(EntityRemoveEvent.Cause.DROP); // CraftBukkit - add Bukkit remove cause
         }
     }
 
