--- a/net/minecraft/world/entity/decoration/Mannequin.java
+++ b/net/minecraft/world/entity/decoration/Mannequin.java
@@ -54,7 +_,7 @@
     private static final String DESCRIPTION_FIELD = "description";
     private static final String HIDE_DESCRIPTION_FIELD = "hide_description";
     private Component description = DEFAULT_DESCRIPTION;
-    private boolean hideDescription = false;
+    public boolean hideDescription = false;
 
     public Mannequin(EntityType<Mannequin> type, Level level) {
         super(type, level);
