--- a/net/minecraft/world/level/chunk/storage/ChunkStorage.java
+++ b/net/minecraft/world/level/chunk/storage/ChunkStorage.java
@@ -38,6 +_,38 @@
         return this.worker.isOldChunkAround(p_223452_, p_223453_);
     }
 
+    // CraftBukkit start
+    private boolean check(net.minecraft.server.level.ServerChunkCache cps, int x, int z) {
+        ChunkPos pos = new ChunkPos(x, z);
+        if (cps != null) {
+            com.google.common.base.Preconditions.checkState(org.bukkit.Bukkit.isPrimaryThread(), "primary thread");
+            if (cps.hasChunk(x, z)) {
+                return true;
+            }
+        }
+
+        CompoundTag nbt;
+        try {
+            nbt = read(pos).get().orElse(null);
+        } catch (InterruptedException | java.util.concurrent.ExecutionException ex) {
+            throw new RuntimeException(ex);
+        }
+        if (nbt != null) {
+            CompoundTag level = nbt.getCompound("Level");
+            if (level.getBoolean("TerrainPopulated")) {
+                return true;
+            }
+
+            net.minecraft.world.level.chunk.status.ChunkStatus status = net.minecraft.world.level.chunk.status.ChunkStatus.byName(level.getString("Status"));
+            if (status != null && status.isOrAfter(net.minecraft.world.level.chunk.status.ChunkStatus.FEATURES)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+    // CraftBukkit end
+    public @Nullable net.minecraft.server.level.ServerLevel generatoraccess; // CraftBukkit
     public CompoundTag upgradeChunkTag(
         ResourceKey<Level> p_188289_,
         Supplier<DimensionDataStorage> p_188290_,
@@ -57,8 +_,22 @@
                     }
                 }
 
+                // Spigot start - SPIGOT-6806: Quick and dirty way to prevent below zero generation in old chunks, by setting the status to heightmap instead of empty
+                boolean stopBelowZero = false;
+                boolean belowZeroGenerationInExistingChunks = (generatoraccess != null) ? generatoraccess.spigotConfig.belowZeroGenerationInExistingChunks : org.spigotmc.SpigotConfig.belowZeroGenerationInExistingChunks;
+                generatoraccess = null;
+                if (i <= 2730 && !belowZeroGenerationInExistingChunks) {
+                    stopBelowZero = "full".equals(p_188291_.getCompound("Level").getString("Status"));
+                }
+                // Spigot end
+
                 injectDatafixingContext(p_188291_, p_188289_, p_188292_);
                 p_188291_ = DataFixTypes.CHUNK.updateToCurrentVersion(this.fixerUpper, p_188291_, Math.max(1493, i));
+                // Spigot start
+                if (stopBelowZero) {
+                    p_188291_.putString("Status", net.minecraft.core.registries.BuiltInRegistries.CHUNK_STATUS.getKey(net.minecraft.world.level.chunk.status.ChunkStatus.SPAWN).toString());
+                }
+                // Spigot end
                 removeDatafixingContext(p_188291_);
                 NbtUtils.addCurrentDataVersion(p_188291_);
                 return p_188291_;
@@ -77,7 +_,7 @@
             synchronized (this) {
                 legacystructuredatahandler = this.legacyStructureHandler;
                 if (legacystructuredatahandler == null) {
-                    this.legacyStructureHandler = legacystructuredatahandler = LegacyStructureDataHandler.getLegacyStructureHandler(p_223449_, p_223450_.get());
+                    this.legacyStructureHandler = legacystructuredatahandler = LegacyStructureDataHandler.getLegacyStructureHandlerCB(p_223449_, p_223450_.get());
                 }
             }
         }
