--- a/net/minecraft/world/level/chunk/storage/EntityStorage.java
+++ b/net/minecraft/world/level/chunk/storage/EntityStorage.java
@@ -92,11 +_,27 @@
         } else {
             try (ProblemReporter.ScopedCollector reporter = new ProblemReporter.ScopedCollector(ChunkAccess.problemPath(pos), LOGGER)) {
                 ListTag entities = new ListTag();
+                final java.util.Map<net.minecraft.world.entity.EntityType<?>, Integer> savedEntityCounts = new java.util.HashMap<>(); // Paper - Entity load/save limit per chunk
                 chunk.getEntities().forEach(e -> {
+                    if (!e.canSaveToDisk()) return; // Purpur - Add canSaveToDisk to Entity
+                    // Paper start - Entity load/save limit per chunk
+                    final EntityType<?> entityType = e.getType();
+                    final int saveLimit = this.level.paperConfig().chunks.entityPerChunkSaveLimit.getOrDefault(entityType, -1);
+                    if (saveLimit > -1) {
+                        if (savedEntityCounts.getOrDefault(entityType, 0) >= saveLimit) {
+                            return;
+                        }
+                        savedEntityCounts.merge(entityType, 1, Integer::sum);
+                    }
+                    // Paper end - Entity load/save limit per chunk
                     TagValueOutput output = TagValueOutput.createWithContext(reporter.forChild(e.problemPath()), e.registryAccess());
+                    try {
                     if (e.save(output)) {
                         CompoundTag result = output.buildResult();
                         entities.add(result);
+                    }
+                    } catch (Exception ex) {
+                        LOGGER.error("An Entity type {} has thrown an exception trying to write state. It will not persist. Report this to the mod author", e.getType(), ex);
                     }
                 });
                 CompoundTag chunkTag = NbtUtils.addCurrentDataVersion(new CompoundTag());
