--- a/net/minecraft/world/level/chunk/storage/RegionFileVersion.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFileVersion.java
@@ -1,5 +_,7 @@
 package net.minecraft.world.level.chunk.storage;
 
+import com.github.luben.zstd.ZstdInputStream;
+import com.github.luben.zstd.ZstdOutputStream;
 import com.mojang.logging.LogUtils;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
 import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
@@ -48,6 +_,14 @@
             p_321471_ -> new BufferedOutputStream(new LZ4BlockOutputStream(p_321471_))
         )
     );
+    public static final RegionFileVersion VERSION_ZSTD = register(
+            new RegionFileVersion(
+                    5,
+                    "zstd",
+                    stream -> new FastBufferedInputStream(new ZstdInputStream(stream)),
+                    stream -> new BufferedOutputStream(new ZstdOutputStream(stream))
+            )
+    );
     public static final RegionFileVersion VERSION_CUSTOM = register(new RegionFileVersion(127, null, p_323443_ -> {
         throw new UnsupportedOperationException();
     }, p_323444_ -> {
@@ -61,6 +_,17 @@
     private final RegionFileVersion.StreamWrapper<InputStream> inputWrapper;
     private final RegionFileVersion.StreamWrapper<OutputStream> outputWrapper;
 
+    // Paper start - Configurable region compression format
+    public static RegionFileVersion getCompressionFormat() {
+        return switch (io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.compressionFormat) {
+            case GZIP -> VERSION_GZIP;
+            case ZLIB -> VERSION_DEFLATE;
+            case LZ4 -> VERSION_LZ4;
+            case ZSTD -> VERSION_ZSTD;
+            case NONE -> VERSION_NONE;
+        };
+    }
+    // Paper end - Configurable region compression format
     private RegionFileVersion(
         int p_63752_, @Nullable String p_321638_, RegionFileVersion.StreamWrapper<InputStream> p_63753_, RegionFileVersion.StreamWrapper<OutputStream> p_63754_
     ) {
