--- a/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -46,7 +_,7 @@
     protected final RegionBitmap usedSectors = new RegionBitmap();
 
     public RegionFile(RegionStorageInfo p_326174_, Path p_196950_, Path p_196951_, boolean p_196952_) throws IOException {
-        this(p_326174_, p_196950_, p_196951_, RegionFileVersion.getSelected(), p_196952_);
+        this(p_326174_, p_196950_, p_196951_, RegionFileVersion.getCompressionFormat(), p_196952_);
     }
 
     public RegionFile(RegionStorageInfo p_326221_, Path p_63633_, Path p_63634_, RegionFileVersion p_63635_, boolean p_63636_) throws IOException {
@@ -82,6 +_,14 @@
                     if (l != 0) {
                         int i1 = getSectorNumber(l);
                         int j1 = getNumSectors(l);
+                        // Spigot start
+                        if (j1 == 255) {
+                            // We're maxed out, so we need to read the proper length from the section
+                            ByteBuffer realLen = ByteBuffer.allocate(4);
+                            this.file.read(realLen, i1 * 4096);
+                            j1 = (realLen.getInt(0) + 4) / 4096 + 1;
+                        }
+                        // Spigot end
                         if (i1 < 2) {
                             LOGGER.warn("Region file {} has invalid sector at index: {}; sector {} overlaps with header", p_63633_, k, i1);
                             this.offsets.put(k, 0);
@@ -118,6 +_,13 @@
             int j = getSectorNumber(i);
             int k = getNumSectors(i);
             int l = k * 4096;
+            // Spigot start
+            if (k == 255) {
+                ByteBuffer realLen = ByteBuffer.allocate(4);
+                this.file.read(realLen, j * 4096);
+                k = (realLen.getInt(0) + 4) / 4096 + 1;
+            }
+            // Spigot end
             ByteBuffer bytebuffer = ByteBuffer.allocate(l);
             this.file.read(bytebuffer, (long)(j * 4096));
             bytebuffer.flip();
