--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -31,6 +_,7 @@
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.nbt.NbtUtils;
 import net.minecraft.nbt.ShortTag;
+import net.minecraft.nbt.Tag;
 import net.minecraft.resources.Identifier;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.ProblemReporter;
@@ -91,7 +_,10 @@
     List<SerializableChunkData.SectionData> sectionData,
     List<CompoundTag> entities,
     List<CompoundTag> blockEntities,
-    CompoundTag structureData
+    CompoundTag structureData,
+    @Nullable CompoundTag attachmentData,
+    @Nullable ListTag auxLightData,
+    @Nullable Tag persistentDataContainer // CraftBukkit
 ) {
     private static final Codec<List<SavedTick<Block>>> BLOCK_TICKS_CODEC = SavedTick.codec(BuiltInRegistries.BLOCK.byNameCodec()).listOf();
     private static final Codec<List<SavedTick<Fluid>>> FLUID_TICKS_CODEC = SavedTick.codec(BuiltInRegistries.FLUID.byNameCodec()).listOf();
@@ -107,11 +_,76 @@
     public static final String BLOCK_LIGHT_TAG = "BlockLight";
     public static final String SKY_LIGHT_TAG = "SkyLight";
 
+    // Paper start - guard against serializing mismatching coordinates
+    // TODO Note: This needs to be re-checked each update
+    public static ChunkPos getChunkCoordinate(final CompoundTag chunkData) {
+        final int dataVersion = NbtUtils.getDataVersion(chunkData);
+        if (dataVersion < 2842) { // Level tag is removed after this version
+            final CompoundTag levelData = chunkData.getCompoundOrEmpty("Level");
+            return new ChunkPos(levelData.getIntOr("xPos", 0), levelData.getIntOr("zPos", 0));
+        } else {
+            return new ChunkPos(chunkData.getIntOr("xPos", 0), chunkData.getIntOr("zPos", 0));
+        }
+    }
+    // Paper end - guard against serializing mismatching coordinates
+    // Paper start - Attempt to recalculate regionfile header if it is corrupt
+    // TODO: Check on update
+    public static long getLastWorldSaveTime(final CompoundTag chunkData) {
+        final int dataVersion = NbtUtils.getDataVersion(chunkData);
+        if (dataVersion < 2842) { // Level tag is removed after this version
+            final CompoundTag levelData = chunkData.getCompoundOrEmpty("Level");
+            return levelData.getLongOr("LastUpdate", 0L);
+        } else {
+            return chunkData.getLongOr("LastUpdate", 0L);
+        }
+    }
+    // Paper end - Attempt to recalculate regionfile header if it is corrupt
+
+    // Paper start - Do not let the server load chunks from newer versions
+    private static final int CURRENT_DATA_VERSION = net.minecraft.SharedConstants.getCurrentVersion().dataVersion().version();
+    private static final boolean JUST_CORRUPT_IT = Boolean.getBoolean("Paper.ignoreWorldDataVersion");
+    // Paper end - Do not let the server load chunks from newer versions
+
+    /**
+     * @deprecated Neo: use constructor with additional data instead
+     */
+    @Deprecated
+    SerializableChunkData(
+            PalettedContainerFactory containerFactory,
+            ChunkPos chunkPos,
+            int minSectionY,
+            long lastUpdateTime,
+            long inhabitedTime,
+            ChunkStatus chunkStatus,
+            BlendingData.@Nullable Packed blendingData,
+            @Nullable BelowZeroRetrogen belowZeroRetrogen,
+            UpgradeData upgradeData,
+            long @Nullable [] carvingMask,
+            Map<Heightmap.Types, long[]> heightmaps,
+            ChunkAccess.PackedTicks packedTicks,
+            ShortList[] postProcessingSections,
+            boolean lightCorrect,
+            List<SerializableChunkData.SectionData> sectionData,
+            List<CompoundTag> entities,
+            List<CompoundTag> blockEntities,
+            CompoundTag structureData
+    ) {
+        this(containerFactory, chunkPos, minSectionY, lastUpdateTime, inhabitedTime, chunkStatus, blendingData, belowZeroRetrogen, upgradeData, carvingMask, heightmaps, packedTicks, postProcessingSections, lightCorrect, sectionData, entities, blockEntities, structureData, null, null, null); // CraftBukkit
+    }
+
     public static SerializableChunkData parse(LevelHeightAccessor levelHeight, PalettedContainerFactory containerFactory, CompoundTag chunkData) {
         if (chunkData.getString("Status").isEmpty()) {
             return null;
         }
 
+        // Paper start - Do not let the server load chunks from newer versions
+        chunkData.getInt("DataVersion").ifPresent(dataVersion -> {
+            if (!JUST_CORRUPT_IT && dataVersion > CURRENT_DATA_VERSION) {
+                new RuntimeException("Server attempted to load chunk saved with newer version of minecraft! " + dataVersion + " > " + CURRENT_DATA_VERSION).printStackTrace();
+                System.exit(1);
+            }
+        });
+        // Paper end - Do not let the server load chunks from newer versions
         ChunkPos chunkPos = new ChunkPos(chunkData.getIntOr("xPos", 0), chunkData.getIntOr("zPos", 0));
         long lastUpdateTime = chunkData.getLongOr("LastUpdate", 0L);
         long inhabitedTime = chunkData.getLongOr("InhabitedTime", 0L);
@@ -123,7 +_,7 @@
         long[] carvingMask = chunkData.getLongArray("carving_mask").orElse(null);
         Map<Heightmap.Types, long[]> heightmaps = new EnumMap<>(Heightmap.Types.class);
         chunkData.getCompound("Heightmaps").ifPresent(heightmapsTag -> {
-            for (Heightmap.Types type : status.heightmapsAfter()) {
+            for (Heightmap.Types type : status.getChunkSaveHeightmaps()) {
                 heightmapsTag.getLongArray(type.getSerializationKey()).ifPresent(longs -> heightmaps.put(type, longs));
             }
         });
@@ -151,7 +_,7 @@
         CompoundTag structureData = chunkData.getCompoundOrEmpty("structures");
         ListTag sectionTags = chunkData.getListOrEmpty("sections");
         List<SerializableChunkData.SectionData> sectionData = new ArrayList<>(sectionTags.size());
-        Codec<PalettedContainerRO<Holder<Biome>>> biomesCodec = containerFactory.biomeContainerCodec();
+            Codec<PalettedContainer<Holder<Biome>>> biomesCodec = containerFactory.biomeContainerRWCodec(); // CraftBukkit - read/write
         Codec<PalettedContainer<BlockState>> blockStatesCodec = containerFactory.blockStatesContainerCodec();
 
         for (int i = 0; i < sectionTags.size(); i++) {
@@ -168,7 +_,7 @@
                                 .getOrThrow(SerializableChunkData.ChunkReadException::new)
                         )
                         .orElseGet(containerFactory::createForBlockStates);
-                    PalettedContainerRO<Holder<Biome>> biomes = sectionTag.getCompound("biomes")
+                        PalettedContainer<Holder<Biome>> biomes = sectionTag.getCompound("biomes")
                         .map(
                             container -> biomesCodec.parse(NbtOps.INSTANCE, container)
                                 .promotePartial(msg -> logErrors(chunkPos, y, msg))
@@ -186,6 +_,11 @@
             }
         }
 
+        CompoundTag attachmentData = chunkData.getCompound(net.neoforged.neoforge.attachment.AttachmentHolder.ATTACHMENTS_NBT_KEY).orElse(null);
+        ListTag auxLightData = chunkData.getList(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY).orElse(null);
+            // CraftBukkit - ChunkBukkitValues
+            Tag persistentDataContainer = chunkData.get("ChunkBukkitValues");
+
         return new SerializableChunkData(
             containerFactory,
             chunkPos,
@@ -204,7 +_,10 @@
             sectionData,
             entities,
             blockEntities,
-            structureData
+            structureData,
+            attachmentData,
+            auxLightData,
+            persistentDataContainer
         );
     }
 
@@ -263,6 +_,9 @@
                 postLoadChunk(level, this.entities, this.blockEntities),
                 BlendingData.unpack(this.blendingData)
             );
+            if (this.auxLightData != null) {
+                ((LevelChunk) chunk).getAuxLightManager(chunkPos).deserializeNBT(this.auxLightData);
+            }
         } else {
             ProtoChunkTicks<Block> blockTicks = ProtoChunkTicks.load(this.packedTicks.blocks());
             ProtoChunkTicks<Fluid> fluidTicks = ProtoChunkTicks.load(this.packedTicks.fluids());
@@ -284,7 +_,7 @@
         chunk.setLightCorrect(this.lightCorrect);
         EnumSet<Heightmap.Types> toPrime = EnumSet.noneOf(Heightmap.Types.class);
 
-        for (Heightmap.Types type : chunk.getPersistedStatus().heightmapsAfter()) {
+        for (Heightmap.Types type : chunk.getPersistedStatus().getChunkSaveHeightmaps()) {
             long[] heightmap = this.heightmaps.get(type);
             if (heightmap != null) {
                 chunk.setHeightmap(type, heightmap);
@@ -304,6 +_,16 @@
             }
         }
 
+        if (this.attachmentData != null) {
+            chunk.readAttachmentsFromNBT(level.registryAccess(), this.attachmentData);
+        }
+
+        // CraftBukkit start - load chunk persistent data from nbt - SPIGOT-6814: Already load PDC here to account for 1.17 to 1.18 chunk upgrading.
+        if (persistentDataContainer instanceof CompoundTag persistentCompoundTag) {
+            chunk.persistentDataContainer.putAll(persistentCompoundTag);
+        }
+        // CraftBukkit end
+
         if (chunkType == ChunkType.LEVELCHUNK) {
             return new ImposterProtoChunk((LevelChunk)chunk, false);
         }
@@ -375,7 +_,7 @@
         Map<Heightmap.Types, long[]> heightmaps = new EnumMap<>(Heightmap.Types.class);
 
         for (Entry<Heightmap.Types, Heightmap> entry : chunk.getHeightmaps()) {
-            if (chunk.getPersistedStatus().heightmapsAfter().contains(entry.getKey())) {
+            if (chunk.getPersistedStatus().getChunkSaveHeightmaps().contains(entry.getKey())) {
                 long[] data = entry.getValue().getRawData();
                 heightmaps.put(entry.getKey(), (long[])data.clone());
             }
@@ -386,25 +_,47 @@
             .map(shorts -> shorts != null && !shorts.isEmpty() ? new ShortArrayList(shorts) : null)
             .toArray(ShortList[]::new);
         CompoundTag structureData = packStructureData(StructurePieceSerializationContext.fromLevel(level), pos, chunk.getAllStarts(), chunk.getAllReferences());
-        return new SerializableChunkData(
-            level.palettedContainerFactory(),
-            pos,
-            chunk.getMinSectionY(),
-            level.getGameTime(),
-            chunk.getInhabitedTime(),
-            chunk.getPersistedStatus(),
-            Optionull.map(chunk.getBlendingData(), BlendingData::pack),
-            chunk.getBelowZeroRetrogen(),
-            chunk.getUpgradeData().copy(),
-            carvingMask,
-            heightmaps,
-            ticksForSerialization,
-            postProcessingSections,
-            chunk.isLightCorrect(),
-            sectionData,
-            entities,
-            blockEntities,
-            structureData
+
+        CompoundTag attachmentData = null;
+        try {
+            attachmentData = chunk.writeAttachmentsToNBT(level.registryAccess());
+        } catch (Exception exception) {
+            LOGGER.error("Failed to write chunk attachments. An attachment has likely thrown an exception trying to write state. It will not persist. Report this to the mod author", exception);
+        }
+        ListTag auxLightData = null;
+        if (chunk instanceof LevelChunk levelChunk) {
+            auxLightData = levelChunk.getAuxLightManager(pos).serializeNBT();
+        }
+
+            // CraftBukkit start - store chunk persistent data in nbt
+            Tag persistentDataContainer = null;
+            if (!chunk.persistentDataContainer.isEmpty()) { // SPIGOT-6814: Always save PDC to account for 1.17 to 1.18 chunk upgrading.
+                persistentDataContainer = chunk.persistentDataContainer.toTagCompound();
+            }
+            // CraftBukkit end
+
+            return new SerializableChunkData(
+                level.palettedContainerFactory(),
+                pos,
+                chunk.getMinSectionY(),
+                level.getGameTime(),
+                chunk.getInhabitedTime(),
+                chunk.getPersistedStatus(),
+                Optionull.map(chunk.getBlendingData(), BlendingData::pack),
+                chunk.getBelowZeroRetrogen(),
+                chunk.getUpgradeData().copy(),
+                carvingMask,
+                heightmaps,
+                ticksForSerialization,
+                postProcessingSections,
+                chunk.isLightCorrect(),
+                sectionData,
+                entities,
+                blockEntities,
+                structureData,
+                attachmentData,
+                auxLightData,
+                persistentDataContainer
         );
     }
 
@@ -471,6 +_,19 @@
         this.heightmaps.forEach((type, data) -> heightmapsTag.put(type.getSerializationKey(), new LongArrayTag(data)));
         tag.put("Heightmaps", heightmapsTag);
         tag.put("structures", this.structureData);
+
+        if (attachmentData != null) {
+            tag.put(net.neoforged.neoforge.attachment.AttachmentHolder.ATTACHMENTS_NBT_KEY, attachmentData);
+        }
+        if (auxLightData != null) {
+            tag.put(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY, auxLightData);
+        }
+        // CraftBukkit start - store chunk persistent data in nbt
+        if (persistentDataContainer != null) { // SPIGOT-6814: Always save PDC to account for 1.17 to 1.18 chunk upgrading.
+            tag.put("ChunkBukkitValues", persistentDataContainer);
+        }
+        // CraftBukkit end
+
         return tag;
     }
 
@@ -549,6 +_,12 @@
             } else {
                 StructureStart start = StructureStart.loadStaticStart(context, startsTag.getCompoundOrEmpty(key), seed);
                 if (start != null) {
+                    // CraftBukkit start - load persistent data for structure start
+                    Tag persistentBase = startsTag.getCompoundOrEmpty(key).get("StructureBukkitValues");
+                    if (persistentBase instanceof CompoundTag persistentCompoundTag) {
+                        start.persistentDataContainer.putAll(persistentCompoundTag);
+                    }
+                    // CraftBukkit end
                     outmap.put(startFeature, start);
                 }
             }
