--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -15,7 +_,6 @@
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
-import java.util.Set;
 import java.util.Map.Entry;
 import net.minecraft.Optionull;
 import net.minecraft.core.BlockPos;
@@ -32,6 +_,7 @@
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.nbt.NbtUtils;
 import net.minecraft.nbt.ShortTag;
+import net.minecraft.nbt.Tag;
 import net.minecraft.resources.Identifier;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.util.ProblemReporter;
@@ -45,6 +_,7 @@
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.CarvingMask;
 import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.level.chunk.ChunkSource;
 import net.minecraft.world.level.chunk.DataLayer;
@@ -83,6 +_,7 @@
     BlendingData.@Nullable Packed blendingData,
     @Nullable BelowZeroRetrogen belowZeroRetrogen,
     UpgradeData upgradeData,
+    long @Nullable [] carvingMask,
     Map<Heightmap.Types, long[]> heightmaps,
     ChunkAccess.PackedTicks packedTicks,
     @Nullable ShortList[] postProcessingSections,
@@ -90,7 +_,10 @@
     List<SerializableChunkData.SectionData> sectionData,
     List<CompoundTag> entities,
     List<CompoundTag> blockEntities,
-    CompoundTag structureData
+    CompoundTag structureData,
+    @Nullable CompoundTag attachmentData,
+    @Nullable ListTag auxLightData,
+    @Nullable Tag persistentDataContainer // CraftBukkit
 ) {
     private static final Codec<List<SavedTick<Block>>> BLOCK_TICKS_CODEC = SavedTick.codec(BuiltInRegistries.BLOCK.byNameCodec()).listOf();
     private static final Codec<List<SavedTick<Fluid>>> FLUID_TICKS_CODEC = SavedTick.codec(BuiltInRegistries.FLUID.byNameCodec()).listOf();
@@ -106,11 +_,77 @@
     public static final String BLOCK_LIGHT_TAG = "BlockLight";
     public static final String SKY_LIGHT_TAG = "SkyLight";
 
+    // Paper start - guard against serializing mismatching coordinates
+    // TODO Note: This needs to be re-checked each update
+    public static ChunkPos getChunkCoordinate(final CompoundTag chunkData) {
+        final int dataVersion = NbtUtils.getDataVersion(chunkData);
+        if (dataVersion < 2842) { // Level tag is removed after this version
+            final CompoundTag levelData = chunkData.getCompoundOrEmpty("Level");
+            return new ChunkPos(levelData.getIntOr("xPos", 0), levelData.getIntOr("zPos", 0));
+        } else {
+            return new ChunkPos(chunkData.getIntOr("xPos", 0), chunkData.getIntOr("zPos", 0));
+        }
+    }
+    // Paper end - guard against serializing mismatching coordinates
+    // Paper start - Attempt to recalculate regionfile header if it is corrupt
+    // TODO: Check on update
+    public static long getLastWorldSaveTime(final CompoundTag chunkData) {
+        final int dataVersion = NbtUtils.getDataVersion(chunkData);
+        if (dataVersion < 2842) { // Level tag is removed after this version
+            final CompoundTag levelData = chunkData.getCompoundOrEmpty("Level");
+            return levelData.getLongOr("LastUpdate", 0L);
+        } else {
+            return chunkData.getLongOr("LastUpdate", 0L);
+        }
+    }
+    // Paper end - Attempt to recalculate regionfile header if it is corrupt
+
+    // Paper start - Do not let the server load chunks from newer versions
+    private static final int CURRENT_DATA_VERSION = net.minecraft.SharedConstants.getCurrentVersion().dataVersion().version();
+    private static final boolean JUST_CORRUPT_IT = Boolean.getBoolean("Paper.ignoreWorldDataVersion");
+    // Paper end - Do not let the server load chunks from newer versions
+
+    /**
+     * @deprecated Neo: use constructor with additional data instead
+     */
+    @Deprecated
+    SerializableChunkData(
+            PalettedContainerFactory containerFactory,
+            ChunkPos chunkPos,
+            int minSectionY,
+            long lastUpdateTime,
+            long inhabitedTime,
+            ChunkStatus chunkStatus,
+            BlendingData.@Nullable Packed blendingData,
+            @Nullable BelowZeroRetrogen belowZeroRetrogen,
+            UpgradeData upgradeData,
+            long @Nullable [] carvingMask,
+            Map<Heightmap.Types, long[]> heightmaps,
+            ChunkAccess.PackedTicks packedTicks,
+            ShortList[] postProcessingSections,
+            boolean lightCorrect,
+            List<SerializableChunkData.SectionData> sectionData,
+            List<CompoundTag> entities,
+            List<CompoundTag> blockEntities,
+            CompoundTag structureData
+    ) {
+        this(containerFactory, chunkPos, minSectionY, lastUpdateTime, inhabitedTime, chunkStatus, blendingData, belowZeroRetrogen, upgradeData, carvingMask, heightmaps, packedTicks, postProcessingSections, lightCorrect, sectionData, entities, blockEntities, structureData, null, null, null); // CraftBukkit
+    }
+
     public static SerializableChunkData parse(LevelHeightAccessor levelHeight, PalettedContainerFactory containerFactory, CompoundTag chunkData) {
+        net.minecraft.server.level.ServerLevel serverLevel = (net.minecraft.server.level.ServerLevel) levelHeight; // Paper - Anti-Xray This is is seemingly only called from ChunkMap, where, we have a server level. We'll fight this later if needed.
         if (chunkData.getString("Status").isEmpty()) {
             return null;
         }
 
+        // Paper start - Do not let the server load chunks from newer versions
+        chunkData.getInt("DataVersion").ifPresent(dataVersion -> {
+            if (!JUST_CORRUPT_IT && dataVersion > CURRENT_DATA_VERSION) {
+                new RuntimeException("Server attempted to load chunk saved with newer version of minecraft! " + dataVersion + " > " + CURRENT_DATA_VERSION).printStackTrace();
+                System.exit(1);
+            }
+        });
+        // Paper end - Do not let the server load chunks from newer versions
         ChunkPos chunkPos = new ChunkPos(chunkData.getIntOr("xPos", 0), chunkData.getIntOr("zPos", 0));
         long lastUpdateTime = chunkData.getLongOr("LastUpdate", 0L);
         long inhabitedTime = chunkData.getLongOr("InhabitedTime", 0L);
@@ -119,9 +_,10 @@
         boolean lightCorrect = chunkData.getBooleanOr("isLightOn", false);
         BlendingData.Packed blendingData = chunkData.read("blending_data", BlendingData.Packed.CODEC).orElse(null);
         BelowZeroRetrogen belowZeroRetrogen = chunkData.read("below_zero_retrogen", BelowZeroRetrogen.CODEC).orElse(null);
+        long[] carvingMask = chunkData.getLongArray("carving_mask").orElse(null);
         Map<Heightmap.Types, long[]> heightmaps = new EnumMap<>(Heightmap.Types.class);
         chunkData.getCompound("Heightmaps").ifPresent(heightmapsTag -> {
-            for (Heightmap.Types type : Heightmap.Types.values()) {
+            for (Heightmap.Types type : status.getChunkSaveHeightmaps()) {
                 heightmapsTag.getLongArray(type.getSerializationKey()).ifPresent(longs -> heightmaps.put(type, longs));
             }
         });
@@ -149,7 +_,7 @@
         CompoundTag structureData = chunkData.getCompoundOrEmpty("structures");
         ListTag sectionTags = chunkData.getListOrEmpty("sections");
         List<SerializableChunkData.SectionData> sectionData = new ArrayList<>(sectionTags.size());
-        Codec<PalettedContainerRO<Holder<Biome>>> biomesCodec = containerFactory.biomeContainerCodec();
+        Codec<PalettedContainer<Holder<Biome>>> biomesCodec = containerFactory.biomeContainerRWCodec(); // CraftBukkit - read/write
         Codec<PalettedContainer<BlockState>> blockStatesCodec = containerFactory.blockStatesContainerCodec();
 
         for (int i = 0; i < sectionTags.size(); i++) {
@@ -159,14 +_,16 @@
                 int y = sectionTag.getByteOr("Y", (byte)0);
                 LevelChunkSection section;
                 if (y >= levelHeight.getMinSectionY() && y <= levelHeight.getMaxSectionY()) {
+                    final BlockState[] presetBlockStates = serverLevel.chunkPacketBlockController.getPresetBlockStates(serverLevel, chunkPos, y); // Paper - Anti-Xray - Add preset block states
+                    final Codec<PalettedContainer<BlockState>> antiXrayBlockStateCodec = presetBlockStates == null ? blockStatesCodec : PalettedContainer.codecRW(BlockState.CODEC, containerFactory.blockStatesStrategy(), net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(), presetBlockStates); // Paper - Anti-Xray
                     PalettedContainer<BlockState> blocks = sectionTag.getCompound("block_states")
                         .map(
-                            container -> blockStatesCodec.parse(NbtOps.INSTANCE, container)
+                            container -> antiXrayBlockStateCodec.parse(NbtOps.INSTANCE, container) // Paper - Anti-Xray
                                 .promotePartial(msg -> logErrors(chunkPos, y, msg))
                                 .getOrThrow(SerializableChunkData.ChunkReadException::new)
                         )
                         .orElseGet(containerFactory::createForBlockStates);
-                    PalettedContainerRO<Holder<Biome>> biomes = sectionTag.getCompound("biomes")
+                    PalettedContainer<Holder<Biome>> biomes = sectionTag.getCompound("biomes") // CraftBukkit - read/write
                         .map(
                             container -> biomesCodec.parse(NbtOps.INSTANCE, container)
                                 .promotePartial(msg -> logErrors(chunkPos, y, msg))
@@ -180,9 +_,22 @@
 
                 DataLayer blockLight = sectionTag.getByteArray("BlockLight").map(DataLayer::new).orElse(null);
                 DataLayer skyLight = sectionTag.getByteArray("SkyLight").map(DataLayer::new).orElse(null);
-                sectionData.add(new SerializableChunkData.SectionData(y, section, blockLight, skyLight));
+                // Paper start - starlight
+                SerializableChunkData.SectionData serializableChunkData = new SerializableChunkData.SectionData(y, section, blockLight, skyLight);
+                if (sectionTag.contains(ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.BLOCKLIGHT_STATE_TAG)) {
+                    ((ca.spottedleaf.moonrise.patches.starlight.storage.StarlightSectionData)(Object)serializableChunkData).starlight$setBlockLightState(sectionTag.getIntOr(ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.BLOCKLIGHT_STATE_TAG, 0));
+                }
+
+                if (sectionTag.contains(ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.SKYLIGHT_STATE_TAG)) {
+                    ((ca.spottedleaf.moonrise.patches.starlight.storage.StarlightSectionData)(Object)serializableChunkData).starlight$setSkyLightState(sectionTag.getIntOr(ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.SKYLIGHT_STATE_TAG, 0));
+                }
+                sectionData.add(serializableChunkData);
+                // Paper end - starlight
             }
         }
+
+        CompoundTag attachmentData = chunkData.getCompound(net.neoforged.neoforge.attachment.AttachmentHolder.ATTACHMENTS_NBT_KEY).orElse(null);
+        ListTag auxLightData = chunkData.getList(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY).orElse(null);
 
         return new SerializableChunkData(
             containerFactory,
@@ -194,6 +_,7 @@
             blendingData,
             belowZeroRetrogen,
             upgradeData,
+            carvingMask,
             heightmaps,
             packedTicks,
             postProcessingSections,
@@ -201,9 +_,65 @@
             sectionData,
             entities,
             blockEntities,
-            structureData
+            structureData,
+            attachmentData,
+            auxLightData,
+            chunkData.get("ChunkBukkitValues")
         );
     }
+
+    // Paper start - starlight
+    private ProtoChunk loadStarlightLightData(final ServerLevel world, final ProtoChunk ret) {
+
+        final boolean hasSkyLight = world.dimensionType().hasSkyLight();
+        final int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinLightSection(world);
+
+        final ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray[] blockNibbles = ca.spottedleaf.moonrise.patches.starlight.light.StarLightEngine.getFilledEmptyLight(world);
+        final ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray[] skyNibbles = ca.spottedleaf.moonrise.patches.starlight.light.StarLightEngine.getFilledEmptyLight(world);
+
+        if (!this.lightCorrect) {
+            ((ca.spottedleaf.moonrise.patches.starlight.chunk.StarlightChunk)ret).starlight$setBlockNibbles(blockNibbles);
+            ((ca.spottedleaf.moonrise.patches.starlight.chunk.StarlightChunk)ret).starlight$setSkyNibbles(skyNibbles);
+            return ret;
+        }
+
+        try {
+            for (final SerializableChunkData.SectionData sectionData : this.sectionData) {
+                final int y = sectionData.y();
+                final DataLayer blockLight = sectionData.blockLight();
+                final DataLayer skyLight = sectionData.skyLight();
+
+                final int blockState = ((ca.spottedleaf.moonrise.patches.starlight.storage.StarlightSectionData)(Object)sectionData).starlight$getBlockLightState();
+                final int skyState = ((ca.spottedleaf.moonrise.patches.starlight.storage.StarlightSectionData)(Object)sectionData).starlight$getSkyLightState();
+
+                if (blockState >= 0) {
+                    if (blockLight != null) {
+                        blockNibbles[y - minSection] = new ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray(ca.spottedleaf.moonrise.common.util.MixinWorkarounds.clone(blockLight.getData()), blockState); // clone for data safety
+                    } else {
+                        blockNibbles[y - minSection] = new ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray(null, blockState);
+                    }
+                }
+
+                if (skyState >= 0 && hasSkyLight) {
+                    if (skyLight != null) {
+                        skyNibbles[y - minSection] = new ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray(ca.spottedleaf.moonrise.common.util.MixinWorkarounds.clone(skyLight.getData()), skyState); // clone for data safety
+                    } else {
+                        skyNibbles[y - minSection] = new ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray(null, skyState);
+                    }
+                }
+            }
+
+            ((ca.spottedleaf.moonrise.patches.starlight.chunk.StarlightChunk)ret).starlight$setBlockNibbles(blockNibbles);
+            ((ca.spottedleaf.moonrise.patches.starlight.chunk.StarlightChunk)ret).starlight$setSkyNibbles(skyNibbles);
+        } catch (final Throwable thr) {
+            ret.setLightCorrect(false);
+
+            LOGGER.error("Failed to parse light data for chunk " + ret.getPos() + " in world '" + ca.spottedleaf.moonrise.common.util.WorldUtil.getWorldName(world) + "'", thr);
+        }
+
+        return ret;
+    }
+    // Paper end - starlight
 
     public ProtoChunk read(ServerLevel level, PoiManager poiManager, RegionStorageInfo regionInfo, ChunkPos pos) {
         if (!Objects.equals(pos, this.chunkPos)) {
@@ -223,7 +_,7 @@
             SectionPos sectionPos = SectionPos.of(pos, section.y);
             if (section.chunkSection != null) {
                 sections[level.getSectionIndexFromSectionY(section.y)] = section.chunkSection;
-                poiManager.checkConsistencyWithBlocks(sectionPos, section.chunkSection);
+                //poiManager.checkConsistencyWithBlocks(sectionPos, section.chunkSection); // Paper - rewrite chunk system
             }
 
             boolean hasBlockLight = section.blockLight != null;
@@ -260,6 +_,9 @@
                 postLoadChunk(level, this.entities, this.blockEntities),
                 BlendingData.unpack(this.blendingData)
             );
+            if (this.auxLightData != null) {
+                ((LevelChunk) chunk).getAuxLightManager(chunkPos).deserializeNBT(this.auxLightData);
+            }
         } else {
             ProtoChunkTicks<Block> blockTicks = ProtoChunkTicks.load(this.packedTicks.blocks());
             ProtoChunkTicks<Fluid> fluidTicks = ProtoChunkTicks.load(this.packedTicks.fluids());
@@ -278,12 +_,22 @@
             }
         }
 
+        // CraftBukkit start - load chunk persistent data from nbt - SPIGOT-6814: Already load PDC here to account for 1.17 to 1.18 chunk upgrading.
+        if (this.persistentDataContainer instanceof CompoundTag compoundTag) {
+            chunk.persistentDataContainer.putAll(compoundTag);
+        }
+        // CraftBukkit end
+
         chunk.setLightCorrect(this.lightCorrect);
-        Set<Heightmap.Types> toPrime = EnumSet.copyOf(chunk.getPersistedStatus().heightmapsAfter());
+        EnumSet<Heightmap.Types> toPrime = EnumSet.noneOf(Heightmap.Types.class);
 
-        for (Entry<Heightmap.Types, long[]> entry : this.heightmaps.entrySet()) {
-            chunk.setHeightmap(entry.getKey(), entry.getValue());
-            toPrime.remove(entry.getKey());
+        for (Heightmap.Types type : chunk.getPersistedStatus().getChunkSaveHeightmaps()) {
+            long[] heightmap = this.heightmaps.get(type);
+            if (heightmap != null) {
+                chunk.setHeightmap(type, heightmap);
+            } else {
+                toPrime.add(type);
+            }
         }
 
         Heightmap.primeHeightmaps(chunk, toPrime);
@@ -297,6 +_,10 @@
             }
         }
 
+        if (this.attachmentData != null) {
+            chunk.readAttachmentsFromNBT(level.registryAccess(), this.attachmentData);
+        }
+
         if (chunkType == ChunkType.LEVELCHUNK) {
             return new ImposterProtoChunk((LevelChunk)chunk, false);
         }
@@ -311,11 +_,15 @@
             protoChunk.setBlockEntityNbt(blockEntity);
         }
 
-        return protoChunk;
+        if (this.carvingMask != null) {
+            protoChunk.setCarvingMask(new CarvingMask(this.carvingMask, chunk.getMinY()));
+        }
+
+        return this.loadStarlightLightData(level, protoChunk); // Paper - starlight
     }
 
     private static void logErrors(ChunkPos pos, int sectionY, String message) {
-        LOGGER.error("Recoverable errors when loading section [{}, {}, {}]: {}", pos.x(), sectionY, pos.z(), message);
+        LOGGER.debug("Recoverable errors when loading section [{}, {}, {}]: {}", pos.x(), sectionY, pos.z(), message);
     }
 
     public static SerializableChunkData copyOf(ServerLevel level, ChunkAccess chunk) {
@@ -325,21 +_,44 @@
 
         ChunkPos pos = chunk.getPos();
         List<SerializableChunkData.SectionData> sectionData = new ArrayList<>();
-        LevelChunkSection[] chunkSections = chunk.getSections();
-        LevelLightEngine lightEngine = level.getChunkSource().getLightEngine();
-
-        for (int sectionY = lightEngine.getMinLightSection(); sectionY < lightEngine.getMaxLightSection(); sectionY++) {
-            int sectionIndex = chunk.getSectionIndexFromSectionY(sectionY);
-            boolean hasSection = sectionIndex >= 0 && sectionIndex < chunkSections.length;
-            DataLayer sourceBlockLight = lightEngine.getLayerListener(LightLayer.BLOCK).getDataLayerData(SectionPos.of(pos, sectionY));
-            DataLayer sourceSkyLight = lightEngine.getLayerListener(LightLayer.SKY).getDataLayerData(SectionPos.of(pos, sectionY));
-            DataLayer blockLight = sourceBlockLight != null && !sourceBlockLight.isEmpty() ? sourceBlockLight.copy() : null;
-            DataLayer skyLight = sourceSkyLight != null && !sourceSkyLight.isEmpty() ? sourceSkyLight.copy() : null;
-            if (hasSection || blockLight != null || skyLight != null) {
-                LevelChunkSection section = hasSection ? chunkSections[sectionIndex].copy() : null;
-                sectionData.add(new SerializableChunkData.SectionData(sectionY, section, blockLight, skyLight));
-            }
+        // Paper start - starlight
+        final int minLightSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinLightSection(level);
+        final int maxLightSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMaxLightSection(level);
+        final int minBlockSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection(level);
+
+        final LevelChunkSection[] chunkSections = chunk.getSections();
+        final ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray[] blockNibbles = ((ca.spottedleaf.moonrise.patches.starlight.chunk.StarlightChunk) chunk).starlight$getBlockNibbles();
+        final ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray[] skyNibbles = ((ca.spottedleaf.moonrise.patches.starlight.chunk.StarlightChunk) chunk).starlight$getSkyNibbles();
+
+        for (int lightSection = minLightSection; lightSection <= maxLightSection; ++lightSection) {
+            final int lightSectionIdx = lightSection - minLightSection;
+            final int blockSectionIdx = lightSection - minBlockSection;
+
+            final LevelChunkSection chunkSection = (blockSectionIdx >= 0 && blockSectionIdx < chunkSections.length) ? chunkSections[blockSectionIdx].copy() : null;
+            final ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray.SaveState blockNibble = blockNibbles[lightSectionIdx].getSaveState();
+            final ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray.SaveState skyNibble = skyNibbles[lightSectionIdx].getSaveState();
+
+            if (chunkSection == null && blockNibble == null && skyNibble == null) {
+                continue;
+            }
+
+            final SerializableChunkData.SectionData section = new SerializableChunkData.SectionData(
+                    lightSection, chunkSection,
+                    blockNibble == null ? null : (blockNibble.data == null ? null : new DataLayer(blockNibble.data)),
+                    skyNibble == null ? null : (skyNibble.data == null ? null : new DataLayer(skyNibble.data))
+            );
+
+            if (blockNibble != null) {
+                ((ca.spottedleaf.moonrise.patches.starlight.storage.StarlightSectionData) (Object) section).starlight$setBlockLightState(blockNibble.state);
+            }
+
+            if (skyNibble != null) {
+                ((ca.spottedleaf.moonrise.patches.starlight.storage.StarlightSectionData) (Object) section).starlight$setSkyLightState(skyNibble.state);
+            }
+
+            sectionData.add(section);
         }
+        // Paper end - starlight
 
         List<CompoundTag> blockEntities = new ArrayList<>(chunk.getBlockEntitiesPos().size());
 
@@ -351,41 +_,71 @@
         }
 
         List<CompoundTag> entities = new ArrayList<>();
+        long[] carvingMask = null;
         if (chunk.getPersistedStatus().getChunkType() == ChunkType.PROTOCHUNK) {
-            ProtoChunk protoChunk = (ProtoChunk)chunk;
+            ProtoChunk protoChunk = (ProtoChunk) chunk;
             entities.addAll(protoChunk.getEntities());
+            CarvingMask existingMask = protoChunk.getCarvingMask();
+            if (existingMask != null) {
+                carvingMask = existingMask.toArray();
+            }
         }
 
         Map<Heightmap.Types, long[]> heightmaps = new EnumMap<>(Heightmap.Types.class);
 
         for (Entry<Heightmap.Types, Heightmap> entry : chunk.getHeightmaps()) {
-            long[] data = entry.getValue().getRawData();
-            heightmaps.put(entry.getKey(), (long[])data.clone());
+            if (chunk.getPersistedStatus().getChunkSaveHeightmaps().contains(entry.getKey())) {
+                long[] data = entry.getValue().getRawData();
+                heightmaps.put(entry.getKey(), (long[]) data.clone());
+            }
         }
 
         ChunkAccess.PackedTicks ticksForSerialization = chunk.getTicksForSerialization(level.getGameTime());
         ShortList[] postProcessingSections = Arrays.stream(chunk.getPostProcessing())
-            .map(shorts -> shorts != null && !shorts.isEmpty() ? new ShortArrayList(shorts) : null)
-            .toArray(ShortList[]::new);
+                .map(shorts -> shorts != null && !shorts.isEmpty() ? new ShortArrayList(shorts) : null)
+                .toArray(ShortList[]::new);
         CompoundTag structureData = packStructureData(StructurePieceSerializationContext.fromLevel(level), pos, chunk.getAllStarts(), chunk.getAllReferences());
+
+        CompoundTag attachmentData = null;
+        try {
+            attachmentData = chunk.writeAttachmentsToNBT(level.registryAccess());
+        } catch (Exception exception) {
+            LOGGER.error("Failed to write chunk attachments. An attachment has likely thrown an exception trying to write state. It will not persist. Report this to the mod author", exception);
+        }
+        ListTag auxLightData = null;
+        if (chunk instanceof LevelChunk levelChunk) {
+            auxLightData = levelChunk.getAuxLightManager(pos).serializeNBT();
+        }
+
+        // CraftBukkit start - store chunk persistent data in nbt
+        CompoundTag persistentDataContainer = null;
+        if (!chunk.persistentDataContainer.isEmpty()) { // SPIGOT-6814: Always save PDC to account for 1.17 to 1.18 chunk upgrading.
+            persistentDataContainer = chunk.persistentDataContainer.toTagCompound();
+        }
+        // CraftBukkit end
+
         return new SerializableChunkData(
-            level.palettedContainerFactory(),
-            pos,
-            chunk.getMinSectionY(),
-            level.getGameTime(),
-            chunk.getInhabitedTime(),
-            chunk.getPersistedStatus(),
-            Optionull.map(chunk.getBlendingData(), BlendingData::pack),
-            chunk.getBelowZeroRetrogen(),
-            chunk.getUpgradeData().copy(),
-            heightmaps,
-            ticksForSerialization,
-            postProcessingSections,
-            chunk.isLightCorrect(),
-            sectionData,
-            entities,
-            blockEntities,
-            structureData
+                level.palettedContainerFactory(),
+                pos,
+                chunk.getMinSectionY(),
+                level.getGameTime(),
+                chunk.getInhabitedTime(),
+                chunk.getPersistedStatus(),
+                Optionull.map(chunk.getBlendingData(), BlendingData::pack),
+                chunk.getBelowZeroRetrogen(),
+                chunk.getUpgradeData().copy(),
+                carvingMask,
+                heightmaps,
+                ticksForSerialization,
+                postProcessingSections,
+                chunk.isLightCorrect(),
+                sectionData,
+                entities,
+                blockEntities,
+                structureData,
+                attachmentData,
+                auxLightData,
+                persistentDataContainer
         );
     }
 
@@ -423,6 +_,19 @@
                 sectionTag.putByteArray("SkyLight", section.skyLight.getData());
             }
 
+            // Paper start - starlight
+            final int blockState = ((ca.spottedleaf.moonrise.patches.starlight.storage.StarlightSectionData)(Object)section).starlight$getBlockLightState();
+            final int skyState = ((ca.spottedleaf.moonrise.patches.starlight.storage.StarlightSectionData)(Object)section).starlight$getSkyLightState();
+
+            if (blockState > 0) {
+                sectionTag.putInt(ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.BLOCKLIGHT_STATE_TAG, blockState);
+            }
+
+            if (skyState > 0) {
+                sectionTag.putInt(ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.SKYLIGHT_STATE_TAG, skyState);
+            }
+            // Paper end - starlight
+
             if (!sectionTag.isEmpty()) {
                 sectionTag.putByte("Y", (byte)section.y);
                 sectionTags.add(sectionTag);
@@ -441,6 +_,9 @@
             ListTag entityTags = new ListTag();
             entityTags.addAll(this.entities);
             tag.put("entities", entityTags);
+            if (this.carvingMask != null) {
+                tag.putLongArray("carving_mask", this.carvingMask);
+            }
         }
 
         saveTicks(tag, this.packedTicks);
@@ -449,6 +_,26 @@
         this.heightmaps.forEach((type, data) -> heightmapsTag.put(type.getSerializationKey(), new LongArrayTag(data)));
         tag.put("Heightmaps", heightmapsTag);
         tag.put("structures", this.structureData);
+
+        if (attachmentData != null) {
+            tag.put(net.neoforged.neoforge.attachment.AttachmentHolder.ATTACHMENTS_NBT_KEY, attachmentData);
+        }
+        if (auxLightData != null) {
+            tag.put(net.neoforged.neoforge.common.world.LevelChunkAuxiliaryLightManager.LIGHT_NBT_KEY, auxLightData);
+        }
+        // CraftBukkit start - store chunk persistent data in nbt
+        if (this.persistentDataContainer != null) { // SPIGOT-6814: Always save PDC to account for 1.17 to 1.18 chunk upgrading.
+            tag.put("ChunkBukkitValues", this.persistentDataContainer);
+        }
+        // CraftBukkit end
+        // Paper start - starlight
+        if (this.lightCorrect && !this.chunkStatus.isBefore(net.minecraft.world.level.chunk.status.ChunkStatus.LIGHT)) {
+            // clobber vanilla value to force vanilla to relight
+            tag.putBoolean("isLightOn", false);
+            // store our light version
+            tag.putInt(ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.STARLIGHT_VERSION_TAG, ca.spottedleaf.moonrise.patches.starlight.util.SaveUtil.STARLIGHT_LIGHT_VERSION);
+        }
+        // Paper end - starlight
         return tag;
     }
 
@@ -527,6 +_,12 @@
             } else {
                 StructureStart start = StructureStart.loadStaticStart(context, startsTag.getCompoundOrEmpty(key), seed);
                 if (start != null) {
+                    // CraftBukkit start - load persistent data for structure start
+                    Tag persistentBase = startsTag.getCompoundOrEmpty(key).get("StructureBukkitValues");
+                    if (persistentBase instanceof CompoundTag persistentCompoundTag) {
+                        start.persistentDataContainer.putAll(persistentCompoundTag);
+                    }
+                    // CraftBukkit end
                     outmap.put(startFeature, start);
                 }
             }
@@ -585,6 +_,60 @@
         }
     }
 
-    public record SectionData(int y, @Nullable LevelChunkSection chunkSection, @Nullable DataLayer blockLight, @Nullable DataLayer skyLight) {
+    // Paper start - starlight - convert from record
+    public static final class SectionData implements ca.spottedleaf.moonrise.patches.starlight.storage.StarlightSectionData { // Paper - starlight - our diff
+        private final int y;
+        private final @Nullable LevelChunkSection chunkSection;
+        private final @Nullable DataLayer blockLight;
+        private final @Nullable DataLayer skyLight;
+
+        // Paper start - starlight - our diff
+        private int blockLightState = -1;
+        private int skyLightState = -1;
+
+        @Override
+        public final int starlight$getBlockLightState() {
+            return this.blockLightState;
+        }
+
+        @Override
+        public final void starlight$setBlockLightState(final int state) {
+            this.blockLightState = state;
+        }
+
+        @Override
+        public final int starlight$getSkyLightState() {
+            return this.skyLightState;
+        }
+
+        @Override
+        public final void starlight$setSkyLightState(final int state) {
+            this.skyLightState = state;
+        }
+        // Paper end - starlight - our diff
+
+        public SectionData(int y, @Nullable LevelChunkSection chunkSection, @Nullable DataLayer blockLight, @Nullable DataLayer skyLight) {
+            this.y = y;
+            this.chunkSection = chunkSection;
+            this.blockLight = blockLight;
+            this.skyLight = skyLight;
+        }
+
+        public int y() {
+            return this.y;
+        }
+
+        public @Nullable LevelChunkSection chunkSection() {
+            return this.chunkSection;
+        }
+
+        public @Nullable DataLayer blockLight() {
+            return this.blockLight;
+        }
+
+        public @Nullable DataLayer skyLight() {
+            return this.skyLight;
+        }
+        // Paper end - starlight - convert from record
     }
 }
