--- a/net/minecraft/recipebook/ServerPlaceRecipe.java
+++ b/net/minecraft/recipebook/ServerPlaceRecipe.java
@@ -31,6 +_,7 @@
             this.inventory = p_135435_.getInventory();
             if (this.testClearGrid() || p_135435_.isCreative()) {
                 this.stackedContents.clear();
+                this.stackedContents.initializeExtras(p_301150_.value(), null); // Paper - Improve exact choice recipe ingredients
                 p_135435_.getInventory().fillStackedContents(this.stackedContents);
                 this.menu.fillCraftSlotsStackedContents(this.stackedContents);
                 if (this.stackedContents.canCraft(p_301150_.value(), null)) {
