--- a/com/mojang/blaze3d/platform/MonitorManager.java
+++ b/com/mojang/blaze3d/platform/MonitorManager.java
@@ -8,7 +_,6 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
-import org.lwjgl.sdl.SDLStdinc;
 import org.lwjgl.sdl.SDLVideo;
 import org.slf4j.Logger;
 
@@ -20,12 +_,8 @@
     public MonitorManager() {
         IntBuffer displays = SDLVideo.SDL_GetDisplays();
         if (displays != null) {
-            try {
-                for (int i = 0; i < displays.limit(); i++) {
-                    this.addDisplay(displays.get(i));
-                }
-            } finally {
-                SDLStdinc.SDL_free(displays);
+            for (int i = 0; i < displays.limit(); i++) {
+                this.addDisplay(displays.get(i));
             }
         }
     }
@@ -42,14 +_,6 @@
         RenderSystem.assertOnRenderThread();
         Monitor monitor = this.monitors.remove(id);
         LOGGER.debug("Monitor {} disconnected. Current monitors: {}", monitor, this.monitors);
-    }
-
-    public void onDisplayModeChanged(int id) {
-        RenderSystem.assertOnRenderThread();
-        Monitor monitor = this.addDisplay(id);
-        if (monitor != null) {
-            LOGGER.debug("Monitor {} mode changed. Current monitors: {}", monitor, this.monitors);
-        }
     }
 
     private @Nullable Monitor addDisplay(int id) {
