--- a/com/mojang/blaze3d/platform/NativeLibrariesBootstrap.java
+++ b/com/mojang/blaze3d/platform/NativeLibrariesBootstrap.java
@@ -47,8 +_,8 @@
 import org.lwjgl.system.Platform;
 import org.lwjgl.util.freetype.FreeType;
 import org.lwjgl.util.shaderc.Shaderc;
-import org.lwjgl.util.spng.SPNG;
 import org.lwjgl.util.spvc.Spvc;
+import org.lwjgl.util.tinyfd.TinyFileDialogs;
 import org.lwjgl.util.vma.Vma;
 import org.lwjgl.vulkan.VK;
 import org.slf4j.Logger;
@@ -82,8 +_,8 @@
             entries.add(new NativeLibrariesBootstrap.LibraryLoadEntry("OpenGL", NativeLibrariesBootstrap::loadOpenGL));
             entries.add(new NativeLibrariesBootstrap.LibraryLoadEntry("OpenAL", NativeLibrariesBootstrap::loadOpenAL));
             entries.add(new NativeLibrariesBootstrap.LibraryLoadEntry("STB", NativeLibrariesBootstrap::loadSTB));
+            entries.add(new NativeLibrariesBootstrap.LibraryLoadEntry("tinyfd", NativeLibrariesBootstrap::loadTinyFD));
             entries.add(new NativeLibrariesBootstrap.LibraryLoadEntry("freetype", NativeLibrariesBootstrap::loadFreeType));
-            entries.add(new NativeLibrariesBootstrap.LibraryLoadEntry("SPNG", NativeLibrariesBootstrap::loadSPNG));
             if (vulkanLoaderAvailable) {
                 entries.add(new NativeLibrariesBootstrap.LibraryLoadEntry("shaderc", NativeLibrariesBootstrap::loadShaderc));
                 entries.add(new NativeLibrariesBootstrap.LibraryLoadEntry("spvc", NativeLibrariesBootstrap::loadSpvc));
@@ -320,12 +_,12 @@
         }
     }
 
+    private static void loadTinyFD() {
+        Objects.requireNonNull(TinyFileDialogs.tinyfd_getGlobalChar("tinyfd_version"));
+    }
+
     private static void loadFreeType() {
         Objects.requireNonNull(FreeType.getLibrary());
-    }
-
-    private static void loadSPNG() {
-        Objects.requireNonNull(SPNG.spng_version_string());
     }
 
     private static class CapturingPrintStream extends PrintStream {
