--- a/com/mojang/blaze3d/platform/NativeImage.java
+++ b/com/mojang/blaze3d/platform/NativeImage.java
@@ -6,14 +_,6 @@
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
-import java.lang.foreign.Arena;
-import java.lang.foreign.FunctionDescriptor;
-import java.lang.foreign.Linker;
-import java.lang.foreign.MemorySegment;
-import java.lang.foreign.ValueLayout;
-import java.lang.invoke.MethodHandle;
-import java.lang.invoke.MethodHandles;
-import java.lang.invoke.MethodType;
 import java.nio.ByteBuffer;
 import java.nio.IntBuffer;
 import java.nio.channels.WritableByteChannel;
@@ -27,20 +_,21 @@
 import java.util.function.IntUnaryOperator;
 import net.minecraft.client.gui.font.providers.FreeTypeUtil;
 import net.minecraft.util.ARGB;
+import net.minecraft.util.PngInfo;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 import org.apache.commons.io.IOUtils;
 import org.jspecify.annotations.Nullable;
-import org.lwjgl.PointerBuffer;
+import org.lwjgl.stb.STBIWriteCallback;
+import org.lwjgl.stb.STBImage;
 import org.lwjgl.stb.STBImageResize;
+import org.lwjgl.stb.STBImageWrite;
 import org.lwjgl.system.MemoryStack;
 import org.lwjgl.system.MemoryUtil;
 import org.lwjgl.util.freetype.FT_Bitmap;
 import org.lwjgl.util.freetype.FT_Face;
 import org.lwjgl.util.freetype.FT_GlyphSlot;
 import org.lwjgl.util.freetype.FreeType;
-import org.lwjgl.util.spng.SPNG;
-import org.lwjgl.util.spng.spng_ihdr;
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
@@ -53,6 +_,7 @@
     private final NativeImage.Format format;
     private final int width;
     private final int height;
+    private final boolean useStbFree;
     private long pixels;
     private final long size;
 
@@ -66,6 +_,7 @@
             this.width = width;
             this.height = height;
             this.size = (long)width * height * format.components();
+            this.useStbFree = false;
             if (zero) {
                 this.pixels = MemoryUtil.nmemCalloc(1L, this.size);
             } else {
@@ -81,11 +_,12 @@
         }
     }
 
-    public NativeImage(NativeImage.Format format, int width, int height, long pixels) {
+    public NativeImage(NativeImage.Format format, int width, int height, boolean useStbFree, long pixels) {
         if (width > 0 && height > 0) {
             this.format = format;
             this.width = width;
             this.height = height;
+            this.useStbFree = useStbFree;
             this.pixels = pixels;
             this.size = (long)width * height * format.components();
         } else {
@@ -95,7 +_,7 @@
 
     @Override
     public String toString() {
-        return "NativeImage[" + this.format + " " + this.width + "x" + this.height + "@" + this.pixels + "]";
+        return "NativeImage[" + this.format + " " + this.width + "x" + this.height + "@" + this.pixels + (this.useStbFree ? "S" : "N") + "]";
     }
 
     private boolean isOutsideBounds(int x, int y) {
@@ -103,17 +_,25 @@
     }
 
     public static NativeImage read(InputStream inputStream) throws IOException {
+        return read(NativeImage.Format.RGBA, inputStream);
+    }
+
+    public static NativeImage read(NativeImage.@Nullable Format format, InputStream inputStream) throws IOException {
         ByteBuffer file = null;
 
         try {
             file = TextureUtil.readResource(inputStream);
-            return read(file);
+            return read(format, file);
         } finally {
             MemoryUtil.memFree(file);
             IOUtils.closeQuietly(inputStream);
         }
     }
 
+    public static NativeImage read(ByteBuffer bytes) throws IOException {
+        return read(NativeImage.Format.RGBA, bytes);
+    }
+
     public static NativeImage read(byte[] bytes) throws IOException {
         MemoryStack memoryStack = MemoryStack.stackGet();
         int bytesAvailable = memoryStack.getPointer();
@@ -139,48 +_,29 @@
         return read(nativeBuffer);
     }
 
-    public static NativeImage read(ByteBuffer bytes) throws IOException {
+    public static NativeImage read(NativeImage.@Nullable Format format, ByteBuffer bytes) throws IOException {
+        if (format != null && !format.supportedByStb()) {
+            throw new UnsupportedOperationException("Don't know how to read format " + format);
+        }
+
         if (MemoryUtil.memAddress(bytes) == 0L) {
             throw new IllegalArgumentException("Invalid buffer");
         }
 
-        long context = SPNG.spng_ctx_new(0);
-        if (context == 0L) {
-            throw new IllegalStateException("Failed to create new SPNG context");
-        }
+        PngInfo.validateHeader(bytes);
 
         try (MemoryStack stack = MemoryStack.stackPush()) {
-            SPNG.spng_set_chunk_limits(context, 1048576L, 1048576L);
-            checkSpngError("set buffer", SPNG.spng_set_png_buffer(context, bytes));
-            spng_ihdr header = spng_ihdr.calloc(stack);
-            checkSpngError("read header", SPNG.spng_get_ihdr(context, header));
-            int width = header.width();
-            int height = header.height();
-            PointerBuffer bufferSizePointer = stack.callocPointer(1);
-            checkSpngError("calculate size", SPNG.spng_decoded_image_size(context, 1, bufferSizePointer));
-            int size = Math.toIntExact(bufferSizePointer.get(0));
-            ByteBuffer pixels = null;
-
-            try {
-                pixels = MemoryUtil.memAlloc(size);
-                checkSpngError("decode image", SPNG.spng_decode_image(context, pixels, 1, 1));
-                long address = MemoryUtil.memAddress(pixels);
-                MEMORY_POOL.malloc(address, pixels.limit());
-                pixels = null;
-                return new NativeImage(NativeImage.Format.RGBA, width, height, address);
-            } finally {
-                if (pixels != null) {
-                    MemoryUtil.memFree(pixels);
-                }
+            IntBuffer w = stack.mallocInt(1);
+            IntBuffer h = stack.mallocInt(1);
+            IntBuffer comp = stack.mallocInt(1);
+            ByteBuffer pixels = STBImage.stbi_load_from_memory(bytes, w, h, comp, format == null ? 0 : format.components);
+            if (pixels == null) {
+                throw new IOException("Could not load image: " + STBImage.stbi_failure_reason());
             }
-        } finally {
-            SPNG.spng_ctx_free(context);
-        }
-    }
 
-    private static void checkSpngError(String operation, int result) throws IOException {
-        if (result != 0) {
-            throw new IOException("SPNG operation '" + operation + "' failed: " + SPNG.spng_strerror(result) + " (" + result + ")");
+            long address = MemoryUtil.memAddress(pixels);
+            MEMORY_POOL.malloc(address, pixels.limit());
+            return new NativeImage(format == null ? NativeImage.Format.getStbFormat(comp.get(0)) : format, w.get(0), h.get(0), true, address);
         }
     }
 
@@ -193,7 +_,12 @@
     @Override
     public void close() {
         if (this.pixels != 0L) {
-            MemoryUtil.nmemFree(this.pixels);
+            if (this.useStbFree) {
+                STBImage.nstbi_image_free(this.pixels);
+            } else {
+                MemoryUtil.nmemFree(this.pixels);
+            }
+
             MEMORY_POOL.free(this.pixels);
         }
 
@@ -359,31 +_,36 @@
     }
 
     public void writeToFile(Path file) throws IOException {
+        if (!this.format.supportedByStb()) {
+            throw new UnsupportedOperationException("Don't know how to write format " + this.format);
+        }
+
         this.checkAllocated();
-        long context = SPNG.spng_ctx_new(2);
-
-        try (
-            WritableByteChannel out = Files.newByteChannel(file, OPEN_OPTIONS);
-            Arena arena = Arena.ofConfined();
-            MemoryStack stack = MemoryStack.stackPush();
-        ) {
-            int width = this.getWidth();
-            int height = Math.min(this.getHeight(), Integer.MAX_VALUE / width / this.format.components());
+
+        try (WritableByteChannel out = Files.newByteChannel(file, OPEN_OPTIONS)) {
+            if (!this.writeToChannel(out)) {
+                throw new IOException("Could not write image to the PNG file \"" + file.toAbsolutePath() + "\": " + STBImage.stbi_failure_reason());
+            }
+        }
+    }
+
+    private boolean writeToChannel(WritableByteChannel output) throws IOException {
+        NativeImage.WriteCallback writer = new NativeImage.WriteCallback(output);
+
+        try {
+            int height = Math.min(this.getHeight(), Integer.MAX_VALUE / this.getWidth() / this.format.components());
             if (height < this.getHeight()) {
                 LOGGER.warn("Dropping image height from {} to {} to fit the size into 32-bit signed int", this.getHeight(), height);
             }
 
-            NativeImage.WriteCallback writer = new NativeImage.WriteCallback(out);
-            MemorySegment writerUpcall = writer.createUpcall(arena);
-            checkSpngError("set output", SPNG.nspng_set_png_stream(context, writerUpcall.address(), 0L));
-            spng_ihdr header = spng_ihdr.calloc(stack).width(width).height(height).color_type((byte)this.format.pngColorType).bit_depth((byte)8);
-            checkSpngError("set header", SPNG.spng_set_ihdr(context, header));
-            checkSpngError("write image", SPNG.nspng_encode_image(context, this.pixels, this.size, 256, 2));
+            if (STBImageWrite.nstbi_write_png_to_func(writer.address(), 0L, this.getWidth(), height, this.format.components(), this.pixels, 0) == 0) {
+                return false;
+            }
+
             writer.throwIfException();
-        } catch (IOException e) {
-            throw new IOException("Could not write image to the PNG file \"" + file.toAbsolutePath(), e);
+            return true;
         } finally {
-            SPNG.spng_ctx_free(context);
+            writer.free();
         }
     }
 
@@ -503,10 +_,10 @@
     }
 
     public enum Format {
-        RGBA(4, true, true, true, false, true, 0, 8, 16, 255, 24, 6),
-        RGB(3, true, true, true, false, false, 0, 8, 16, 255, 255, 2),
-        LUMINANCE_ALPHA(2, false, false, false, true, true, 255, 255, 255, 0, 8, 4),
-        LUMINANCE(1, false, false, false, true, false, 0, 0, 0, 0, 255, 0);
+        RGBA(4, true, true, true, false, true, 0, 8, 16, 255, 24, true),
+        RGB(3, true, true, true, false, false, 0, 8, 16, 255, 255, true),
+        LUMINANCE_ALPHA(2, false, false, false, true, true, 255, 255, 255, 0, 8, true),
+        LUMINANCE(1, false, false, false, true, false, 0, 0, 0, 0, 255, true);
 
         private final int components;
         private final boolean hasRed;
@@ -519,7 +_,7 @@
         private final int blueOffset;
         private final int luminanceOffset;
         private final int alphaOffset;
-        private final int pngColorType;
+        private final boolean supportedByStb;
 
         Format(
             int components,
@@ -533,7 +_,7 @@
             int blueOffset,
             int luminanceOffset,
             int alphaOffset,
-            int pngColorType
+            boolean supportedByStb
         ) {
             this.components = components;
             this.hasRed = hasRed;
@@ -546,7 +_,7 @@
             this.blueOffset = blueOffset;
             this.luminanceOffset = luminanceOffset;
             this.alphaOffset = alphaOffset;
-            this.pngColorType = pngColorType;
+            this.supportedByStb = supportedByStb;
         }
 
         public int components() {
@@ -624,47 +_,42 @@
         public int luminanceOrAlphaOffset() {
             return this.hasLuminance ? this.luminanceOffset : this.alphaOffset;
         }
+
+        public boolean supportedByStb() {
+            return this.supportedByStb;
+        }
+
+        private static NativeImage.Format getStbFormat(int i) {
+            switch (i) {
+                case 1:
+                    return LUMINANCE;
+                case 2:
+                    return LUMINANCE_ALPHA;
+                case 3:
+                    return RGB;
+                case 4:
+                default:
+                    return RGBA;
+            }
+        }
     }
 
-    private static class WriteCallback {
-        private static final Linker LINKER = Linker.nativeLinker();
-        private static final FunctionDescriptor CALLBACK_DESC = FunctionDescriptor.of(
-            ValueLayout.JAVA_INT, ValueLayout.ADDRESS, ValueLayout.ADDRESS, ValueLayout.ADDRESS.withTargetLayout(ValueLayout.JAVA_BYTE), ValueLayout.JAVA_LONG
-        );
-        private static final MethodHandle CALLBACK_FN = findCallbackFn();
+    private static class WriteCallback extends STBIWriteCallback {
         private final WritableByteChannel output;
         private @Nullable IOException exception;
 
-        private static MethodHandle findCallbackFn() {
-            try {
-                return MethodHandles.lookup()
-                    .findVirtual(
-                        NativeImage.WriteCallback.class,
-                        "invoke",
-                        MethodType.methodType(int.class, MemorySegment.class, MemorySegment.class, MemorySegment.class, long.class)
-                    );
-            } catch (Exception e) {
-                throw new IllegalStateException(e);
-            }
-        }
-
         private WriteCallback(WritableByteChannel output) {
             this.output = output;
         }
 
-        public MemorySegment createUpcall(Arena arena) {
-            return LINKER.upcallStub(CALLBACK_FN.bindTo(this), CALLBACK_DESC, arena);
-        }
-
-        public int invoke(MemorySegment ctx, MemorySegment user, MemorySegment dest, long length) {
-            ByteBuffer dataBuf = dest.reinterpret(length).asByteBuffer();
+        @Override
+        public void invoke(long context, long data, int size) {
+            ByteBuffer dataBuf = getData(data, size);
 
             try {
                 this.output.write(dataBuf);
-                return 0;
             } catch (IOException e) {
                 this.exception = e;
-                return -2;
             }
         }
 
