--- a/com/mojang/blaze3d/platform/SDLEventHandler.java
+++ b/com/mojang/blaze3d/platform/SDLEventHandler.java
@@ -4,7 +_,6 @@
 import java.util.List;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.gui.screens.options.controls.KeyBindsScreen;
-import net.minecraft.client.input.IMECandidatesEvent;
 import net.minecraft.client.input.InputQuirks;
 import net.minecraft.client.input.KeyEvent;
 import net.minecraft.client.input.MouseButtonInfo;
@@ -18,7 +_,6 @@
 import org.lwjgl.sdl.SDL_MouseButtonEvent;
 import org.lwjgl.sdl.SDL_MouseMotionEvent;
 import org.lwjgl.sdl.SDL_MouseWheelEvent;
-import org.lwjgl.sdl.SDL_TextEditingCandidatesEvent;
 import org.lwjgl.sdl.SDL_TextEditingEvent;
 
 @OnlyIn(Dist.CLIENT)
@@ -53,9 +_,6 @@
                     case 772:
                         this.handleKeymapChangedEvent();
                         break;
-                    case 775:
-                        this.handleTextEditingCandidatesEvent(event);
-                        break;
                     case 1024:
                         this.handleMouseMotionEvent(event);
                         break;
@@ -83,11 +_,6 @@
     }
 
     public void pumpEvents() {
-        this.flushInputEvents();
-        this.pollEvents();
-    }
-
-    public void flushInputEvents() {
         SDLEvents.SDL_PumpEvents();
         SDLEvents.SDL_FlushEvents(768, 4871);
     }
@@ -120,15 +_,6 @@
         PreeditEvent preedit = PreeditEvent.fromSdlTextEditing(edit.textString(), edit.start(), edit.length());
         long handle = getWindowHandle(event);
         this.minecraft.execute(() -> this.minecraft.keyboardHandler.textEditing(handle, preedit));
-    }
-
-    private void handleTextEditingCandidatesEvent(SDL_Event event) {
-        SDL_TextEditingCandidatesEvent candidatesEvent = event.edit_candidates();
-        IMECandidatesEvent candidates = IMECandidatesEvent.fromSdl(
-            candidatesEvent.candidates(), candidatesEvent.num_candidates(), candidatesEvent.selected_candidate(), candidatesEvent.horizontal()
-        );
-        long handle = getWindowHandle(event);
-        this.minecraft.execute(() -> this.minecraft.keyboardHandler.textEditingCandidates(handle, candidates));
     }
 
     private void handleMouseMotionEvent(SDL_Event event) {
