--- a/com/mojang/blaze3d/platform/TextInputManager.java
+++ b/com/mojang/blaze3d/platform/TextInputManager.java
@@ -13,11 +_,6 @@
     private final Window window;
     private boolean textInputEnabled;
     private @Nullable Object owner;
-    private boolean hasTextInputArea;
-    private int areaX;
-    private int areaY;
-    private int areaWidth;
-    private int areaHeight;
 
     public TextInputManager(Window window) {
         this.window = window;
@@ -25,23 +_,13 @@
 
     public void setTextInputArea(int x0, int y0, int x1, int y1) {
         double windowScale = (double)this.window.getGuiScale() / this.window.getPixelDensity();
-        int x = (int)Math.round(x0 * windowScale);
-        int y = (int)Math.round(y0 * windowScale);
-        int width = Math.max(1, (int)Math.round((x1 - x0) * windowScale));
-        int height = Math.max(1, (int)Math.round((y1 - y0) * windowScale));
-        if (!this.hasTextInputArea || x != this.areaX || y != this.areaY || width != this.areaWidth || height != this.areaHeight) {
-            this.areaX = x;
-            this.areaY = y;
-            this.areaWidth = width;
-            this.areaHeight = height;
-            this.hasTextInputArea = true;
-            this.applyTextInputArea();
-        }
-    }
 
-    private void applyTextInputArea() {
         try (MemoryStack stack = MemoryStack.stackPush()) {
-            Buffer rect = SDL_Rect.malloc(1, stack).x(this.areaX).y(this.areaY).w(this.areaWidth).h(this.areaHeight);
+            Buffer rect = SDL_Rect.malloc(1, stack)
+                .x((int)Math.round(x0 * windowScale))
+                .y((int)Math.round(y0 * windowScale))
+                .w(Math.max(1, (int)Math.round((x1 - x0) * windowScale)))
+                .h(Math.max(1, (int)Math.round((y1 - y0) * windowScale)));
             SDLKeyboard.SDL_SetTextInputArea(this.window.handle(), rect, -1);
         }
     }
@@ -49,13 +_,8 @@
     public void startTextInput(Object owner) {
         this.owner = owner;
         if (!this.textInputEnabled) {
-            if (this.hasTextInputArea) {
-                this.applyTextInputArea();
-            }
-
             this.textInputEnabled = true;
             SDLKeyboard.SDL_StartTextInput(this.window.handle());
-            SDLKeyboard.SDL_ClearComposition(this.window.handle());
         }
     }
 
