--- a/net/minecraft/client/renderer/texture/atlas/SpriteSource.java
+++ b/net/minecraft/client/renderer/texture/atlas/SpriteSource.java
@@ -11,10 +_,8 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.extensions.SpriteSourceExtension;
-
 @OnlyIn(Dist.CLIENT)
-public interface SpriteSource extends SpriteSourceExtension {
+public interface SpriteSource {
     FileToIdConverter TEXTURE_ID_CONVERTER = new FileToIdConverter("textures", ".png");
 
     void run(ResourceManager resourceManager, SpriteSource.Output output);
