--- a/com/mojang/renderpearl/backend/api/GpuDeviceBackend.java
+++ b/com/mojang/renderpearl/backend/api/GpuDeviceBackend.java
@@ -4,6 +_,9 @@
 import com.mojang.renderpearl.api.buffers.GpuBuffer;
 import com.mojang.renderpearl.api.commands.GpuQueryPool;
 import com.mojang.renderpearl.api.device.DeviceInfo;
+import com.mojang.renderpearl.api.pipeline.CompiledRenderPipeline;
+import com.mojang.renderpearl.api.pipeline.RenderPipeline;
+import com.mojang.renderpearl.api.pipeline.ShaderSource;
 import com.mojang.renderpearl.api.textures.AddressMode;
 import com.mojang.renderpearl.api.textures.FilterMode;
 import com.mojang.renderpearl.api.textures.GpuSampler;
@@ -12,7 +_,6 @@
 import java.nio.ByteBuffer;
 import java.util.List;
 import java.util.OptionalDouble;
-import java.util.function.BooleanSupplier;
 import java.util.function.Supplier;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
@@ -20,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public interface GpuDeviceBackend {
-    GpuSurfaceBackend createSurface(long windowHandle, BooleanSupplier isIconified);
+    GpuSurfaceBackend createSurface(long windowHandle);
 
     CommandEncoderBackend createCommandEncoder();
 
@@ -28,8 +_,14 @@
         AddressMode addressModeU, AddressMode addressModeV, FilterMode minFilter, FilterMode magFilter, int maxAnisotropy, OptionalDouble maxLod
     );
 
+    GpuTexture createTexture(
+        @Nullable Supplier<String> label, @GpuTexture.Usage int usage, GpuFormat format, int width, int height, int depthOrLayers, int mipLevels
+    );
+
     GpuTexture createTexture(@Nullable String label, @GpuTexture.Usage int usage, GpuFormat format, int width, int height, int depthOrLayers, int mipLevels);
 
+    GpuTextureView createTextureView(GpuTexture texture);
+
     GpuTextureView createTextureView(GpuTexture texture, int baseMipLevel, int mipLevels);
 
     GpuBuffer createBuffer(@Nullable Supplier<String> label, @GpuBuffer.Usage int usage, long size);
@@ -40,13 +_,13 @@
 
     boolean isDebuggingEnabled();
 
-    BackendRenderPipeline.Pending compilePipeline(BackendRenderPipeline.CreateInfo pipelineCreateInfo);
+    @Nullable CompiledRenderPipeline compilePipeline(RenderPipeline pipeline, ShaderSource shaderSource);
 
     void close();
 
     GpuQueryPool createTimestampQueryPool(int size);
 
-    long getTimestampCalibrationOffset();
+    long getTimestampNow();
 
     DeviceInfo getDeviceInfo();
 }
