--- a/com/mojang/renderpearl/backend/api/GpuSurfaceBackend.java
+++ b/com/mojang/renderpearl/backend/api/GpuSurfaceBackend.java
@@ -3,13 +_,12 @@
 import com.mojang.renderpearl.api.device.GpuSurface;
 import com.mojang.renderpearl.api.device.SurfaceException;
 import com.mojang.renderpearl.api.textures.GpuTextureView;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.util.Collection;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public interface GpuSurfaceBackend extends UncheckedAutoCloseable {
+public interface GpuSurfaceBackend extends AutoCloseable {
     void configure(GpuSurface.Configuration config) throws SurfaceException;
 
     boolean isSuboptimal();
@@ -19,6 +_,9 @@
     void blitFromTexture(CommandEncoderBackend commandEncoder, GpuTextureView textureView);
 
     void present();
+
+    @Override
+    void close();
 
     Collection<GpuSurface.PresentMode> supportedPresentModes();
 }
