--- a/com/mojang/renderpearl/backend/api/RenderPassBackend.java
+++ b/com/mojang/renderpearl/backend/api/RenderPassBackend.java
@@ -3,9 +_,13 @@
 import com.mojang.renderpearl.api.buffers.GpuBuffer;
 import com.mojang.renderpearl.api.buffers.GpuBufferSlice;
 import com.mojang.renderpearl.api.commands.GpuQueryPool;
+import com.mojang.renderpearl.api.commands.RenderPass;
+import com.mojang.renderpearl.api.pipeline.CompiledRenderPipeline;
 import com.mojang.renderpearl.api.pipeline.IndexType;
-import java.nio.ByteBuffer;
+import com.mojang.renderpearl.api.textures.GpuSampler;
+import com.mojang.renderpearl.api.textures.GpuTextureView;
 import java.nio.IntBuffer;
+import java.util.Collection;
 import java.util.function.Supplier;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
@@ -18,11 +_,13 @@
 
     void popDebugGroup();
 
-    void setPipeline(final BackendRenderPipeline pipeline);
-
-    void setUniform(final int index, final @Nullable Object value);
-
-    void pushConstants(ByteBuffer value);
+    void setPipeline(final CompiledRenderPipeline pipeline);
+
+    void bindTexture(final String name, final @Nullable GpuTextureView textureView, final @Nullable GpuSampler sampler);
+
+    void setUniform(final String name, final GpuBuffer value);
+
+    void setUniform(final String name, final GpuBufferSlice value);
 
     void enableScissor(final int x, final int y, final int width, final int height);
 
@@ -39,6 +_,14 @@
     void multiDrawIndexed(final PointerBuffer firstIndexOffsets, final IntBuffer indexCounts, final IntBuffer vertexOffsets, final int drawCount);
 
     void drawIndexedIndirect(final GpuBufferSlice commands, final int drawCount);
+
+    <T> void drawMultipleIndexed(
+        final Collection<RenderPass.Draw<T>> draws,
+        final @Nullable GpuBuffer defaultIndexBuffer,
+        final @Nullable IndexType defaultIndexType,
+        final Collection<String> dynamicUniforms,
+        final T uniformArgument
+    );
 
     void draw(final int vertexCount, final int instanceCount, final int firstVertex, final int firstInstance);
 
