--- a/net/minecraft/util/profiling/ActiveProfiler.java
+++ b/net/minecraft/util/profiling/ActiveProfiler.java
@@ -77,6 +_,7 @@
 
     @Override
     public void push(String p_18390_) {
+        if (true) return;
         if (!this.started) {
             LOGGER.error("Cannot push '{}' to profiler if profiler tick hasn't started - missing startTick()?", p_18390_);
         } else {
@@ -103,6 +_,7 @@
 
     @Override
     public void pop() {
+        if (true) return;
         if (!this.started) {
             LOGGER.error("Cannot pop from profiler if profiler tick hasn't started - missing startTick()?");
         } else if (this.startTimes.isEmpty()) {
@@ -132,6 +_,7 @@
 
     @Override
     public void popPush(String p_18395_) {
+        if (true) return;
         this.pop();
         this.push(p_18395_);
     }
