--- a/net/minecraft/util/profiling/ActiveProfiler.java
+++ b/net/minecraft/util/profiling/ActiveProfiler.java
@@ -77,6 +_,7 @@
 
     @Override
     public void push(String name) {
+        if (true) return;
         if (!this.started) {
             LOGGER.error("Cannot push '{}' to profiler if profiler tick hasn't started - missing startTick()?", name);
         } else {
@@ -103,6 +_,7 @@
 
     @Override
     public void pop() {
+        if (true) return;
         if (!this.started) {
             LOGGER.error("Cannot pop from profiler if profiler tick hasn't started - missing startTick()?");
         } else if (this.startTimes.isEmpty()) {
@@ -132,12 +_,14 @@
 
     @Override
     public void popPush(String name) {
+        if (true) return;
         this.pop();
         this.push(name);
     }
 
     @Override
     public void popPush(Supplier<String> name) {
+        if (true) return;
         this.pop();
         this.push(name);
     }
