--- a/net/minecraft/util/profiling/Profiler.java
+++ b/net/minecraft/util/profiling/Profiler.java
@@ -45,6 +_,7 @@
     }
 
     public static ProfilerFiller get() {
+        if (true) return net.minecraft.util.profiling.InactiveProfiler.INSTANCE; // Pufferfish - Option to disable method profiler
         return ACTIVE_COUNT.get() == 0 ? getDefaultFiller() : Objects.requireNonNullElseGet(ACTIVE.get(), Profiler::getDefaultFiller);
     }
 
