--- a/net/minecraft/server/waypoints/ServerWaypointManager.java
+++ b/net/minecraft/server/waypoints/ServerWaypointManager.java
@@ -123,7 +123,7 @@
     }
 
     private static boolean isLocatorBarEnabledFor(EntityPlayer entityplayer) {
-        return entityplayer.level().getServer().getGameRules().getBoolean(GameRules.RULE_LOCATOR_BAR);
+        return entityplayer.level().getGameRules().getBoolean(GameRules.RULE_LOCATOR_BAR); // CraftBukkit - per-world
     }
 
     private void createConnection(EntityPlayer entityplayer, WaypointTransmitter waypointtransmitter) {
