--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -206,7 +206,7 @@
 
     @Override
     public Brain<Creaking> getBrain() {
-        return super.getBrain();
+        return (Brain<Creaking>) super.getBrain(); // CraftBukkit - decompile error
     }
 
     @Override
@@ -329,7 +329,7 @@
         }
 
         this.makeSound(this.getDeathSound());
-        this.remove(Entity.RemovalReason.DISCARDED);
+        this.remove(Entity.RemovalReason.DISCARDED, null); // CraftBukkit - add Bukkit remove cause
     }
 
     public void creakingDeathEffects(DamageSource damagesource) {
@@ -440,7 +440,7 @@
 
     @Override
     protected SoundEvent getHurtSound(DamageSource damagesource) {
-        return this.isHeartBound() ? SoundEvents.CREAKING_SWAY : super.getHurtSound(damagesource);
+        return SoundEvents.CREAKING_SWAY;
     }
 
     @Override
