--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -206,7 +206,7 @@
 
     @Override
     public BehaviorController<Creaking> getBrain() {
-        return super.getBrain();
+        return (BehaviorController<Creaking>) super.getBrain(); // CraftBukkit - decompile error
     }
 
     @Override
@@ -329,7 +329,7 @@
         }
 
         this.makeSound(this.getDeathSound());
-        this.remove(Entity.RemovalReason.DISCARDED);
+        this.remove(Entity.RemovalReason.DISCARDED, null); // CraftBukkit - add Bukkit remove cause
     }
 
     public void creakingDeathEffects(DamageSource damagesource) {
@@ -458,7 +458,7 @@
 
     @Override
     protected SoundEffect getHurtSound(DamageSource damagesource) {
-        return this.isHeartBound() ? SoundEffects.CREAKING_SWAY : super.getHurtSound(damagesource);
+        return SoundEffects.CREAKING_SWAY;
     }
 
     @Override
