--- a/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
+++ b/net/minecraft/commands/synchronization/ArgumentTypeInfos.java
@@ -41,7 +_,6 @@
 import net.minecraft.commands.arguments.SlotSourceArgument;
 import net.minecraft.commands.arguments.SlotsArgument;
 import net.minecraft.commands.arguments.StyleArgument;
-import net.minecraft.commands.arguments.SwingAnimationArgument;
 import net.minecraft.commands.arguments.TeamArgument;
 import net.minecraft.commands.arguments.TeamColorArgument;
 import net.minecraft.commands.arguments.TemplateMirrorArgument;
@@ -69,6 +_,20 @@
 public class ArgumentTypeInfos {
     private static final Map<Class<?>, ArgumentTypeInfo<?, ?>> BY_CLASS = Maps.newHashMap();
 
+    /**
+     * Forge: Use this in conjunction with a
+     * {@link net.neoforged.neoforge.registries.DeferredRegister#register(String, java.util.function.Supplier) DeferredRegister#register(String, Supplier)}
+     * call to both populate the {@code BY_CLASS} map and register the argument type info so it can be used in commands.
+     *
+     * @param infoClass the class type of the argument type info
+     * @param argumentTypeInfo the argument type info instance
+     * @return the provided argument type info instance for chaining
+     */
+    public static synchronized <A extends ArgumentType<?>, T extends ArgumentTypeInfo.Template<A>, I extends ArgumentTypeInfo<A, T>> I registerByClass(Class<A> infoClass, I argumentTypeInfo) {
+        BY_CLASS.put(infoClass, argumentTypeInfo);
+        return argumentTypeInfo;
+    }
+
     private static <A extends ArgumentType<?>, T extends ArgumentTypeInfo.Template<A>> ArgumentTypeInfo<A, T> register(
         Registry<ArgumentTypeInfo<?, ?>> registry, String id, Class<? extends A> brigadierType, ArgumentTypeInfo<A, T> info
     ) {
@@ -139,22 +_,9 @@
         register(
             registry, "loot_modifier", ResourceOrIdArgument.LootModifierArgument.class, SingletonArgumentInfo.contextAware(ResourceOrIdArgument::lootModifier)
         );
-        register(
-            registry,
-            "context_float_provider",
-            ResourceOrIdArgument.ContextFloatProviderArgument.class,
-            SingletonArgumentInfo.contextAware(ResourceOrIdArgument::floatProvider)
-        );
-        register(
-            registry,
-            "context_int_provider",
-            ResourceOrIdArgument.ContextIntProviderArgument.class,
-            SingletonArgumentInfo.contextAware(ResourceOrIdArgument::intProvider)
-        );
         register(registry, "slot_source", SlotSourceArgument.class, SingletonArgumentInfo.contextAware(SlotSourceArgument::slotSource));
         register(registry, "dialog", ResourceOrIdArgument.DialogArgument.class, SingletonArgumentInfo.contextAware(ResourceOrIdArgument::dialog));
         register(registry, "feature", ResourceOrIdArgument.FeatureArgument.class, SingletonArgumentInfo.contextAware(ResourceOrIdArgument::feature));
-        register(registry, "swing_animation", SwingAnimationArgument.class, SingletonArgumentInfo.contextFree(SwingAnimationArgument::swingAnimationType));
         return register(registry, "uuid", UuidArgument.class, SingletonArgumentInfo.contextFree(UuidArgument::uuid));
     }
 
