--- a/net/minecraft/client/model/HumanoidModel.java
+++ b/net/minecraft/client/model/HumanoidModel.java
@@ -28,8 +_,6 @@
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
-import net.neoforged.fml.common.asm.enumextension.IExtensibleEnum;
-
 @OnlyIn(Dist.CLIENT)
 public class HumanoidModel<T extends HumanoidRenderState> extends EntityModel<T> implements ArmedModel<T>, HeadedModel {
     public static final MeshTransformer BABY_TRANSFORMER = new BabyModelTransform(true, 16.0F, 0.0F, 2.0F, 2.0F, 24.0F, Set.of("head"));
@@ -496,7 +_,7 @@
         return this.head;
     }
 
-    public enum ArmPose implements IExtensibleEnum {
+    public enum ArmPose {
         EMPTY(false, false),
         ITEM(false, false),
         BLOCK(false, false),
