--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTrader.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.npc.wanderingtrader;
 
 import java.util.EnumSet;
+import java.util.Objects;
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvent;
@@ -45,6 +_,7 @@
 import net.minecraft.world.level.storage.ValueInput;
 import net.minecraft.world.level.storage.ValueOutput;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.event.entity.EntityRemoveEvent;
 import org.jspecify.annotations.Nullable;
 
 public class WanderingTrader extends AbstractVillager implements Consumable.OverrideConsumeSound {
@@ -54,6 +_,7 @@
 
     public WanderingTrader(EntityType<? extends WanderingTrader> type, Level level) {
         super(type, level);
+        this.setDespawnDelay(48000); // CraftBukkit - set default from MobSpawnerTrader
     }
 
     @Override
@@ -210,6 +_,7 @@
 
     private void maybeDespawn() {
         if (this.despawnDelay > 0 && !this.isTrading() && --this.despawnDelay == 0) {
+            this.removeReason(EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
             this.discard();
         }
     }
@@ -218,7 +_,7 @@
         this.wanderTarget = pos;
     }
 
-    public @Nullable BlockPos getWanderTarget() {
+    private @Nullable BlockPos getWanderTarget() {
         return this.wanderTarget;
     }
 
