--- a/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java
+++ b/net/minecraft/world/entity/npc/wanderingtrader/WanderingTraderSpawner.java
@@ -4,7 +_,6 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.BiomeTags;
-import net.minecraft.util.Continuation;
 import net.minecraft.util.Mth;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.entity.EntitySpawnReason;
@@ -15,6 +_,7 @@
 import net.minecraft.world.entity.ai.village.poi.PoiTypes;
 import net.minecraft.world.entity.animal.equine.TraderLlama;
 import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.CustomSpawner;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.gamerules.GameRules;
@@ -37,30 +_,38 @@
 
     public WanderingTraderSpawner(SavedDataStorage savedDataStorage) {
         this.savedDataStorage = savedDataStorage;
-        this.tickDelay = 1200;
+        this.tickDelay = Integer.MIN_VALUE; // Paper - Add Wandering Trader spawn rate config options
         this.traderData = null;
     }
 
     @Override
-    public void tick(ServerLevel level, boolean spawnEnemies) {
+    public void tick(final ServerLevel level, final boolean spawnEnemies) {
+        // Paper start - Add Wandering Trader spawn rate config options
+        if (this.tickDelay == Integer.MIN_VALUE) {
+            this.tickDelay = level.paperConfig().entities.spawning.wanderingTrader.spawnMinuteLength;
+            WanderingTraderData data = this.getTraderData();
+            data.setSpawnDelay(level.paperConfig().entities.spawning.wanderingTrader.spawnDayLength);
+            data.setSpawnChance(level.paperConfig().entities.spawning.wanderingTrader.spawnChanceMin);
+        }
         if (level.getGameRules().get(GameRules.SPAWN_WANDERING_TRADERS)) {
-            if (--this.tickDelay <= 0) {
-                this.tickDelay = 1200;
+            if (this.tickDelay - 1 <= 0) { // Paper - Prevent tickDelay going below 0
+                this.tickDelay = level.paperConfig().entities.spawning.wanderingTrader.spawnMinuteLength;
                 WanderingTraderData data = this.getTraderData();
-                int spawnDelay = data.spawnDelay() - 1200;
+                int spawnDelay = data.spawnDelay() - level.paperConfig().entities.spawning.wanderingTrader.spawnMinuteLength;
                 data.setSpawnDelay(spawnDelay);
                 if (spawnDelay <= 0) {
-                    data.setSpawnDelay(24000);
+                    data.setSpawnDelay(level.paperConfig().entities.spawning.wanderingTrader.spawnDayLength);
                     int chanceToSpawn = data.spawnChance();
-                    int newSpawnChance = Mth.clamp(chanceToSpawn + 25, 25, 75);
+                    int newSpawnChance = Mth.clamp(chanceToSpawn + level.paperConfig().entities.spawning.wanderingTrader.spawnChanceFailureIncrement, level.paperConfig().entities.spawning.wanderingTrader.spawnChanceMin, level.paperConfig().entities.spawning.wanderingTrader.spawnChanceMax);
                     data.setSpawnChance(newSpawnChance);
                     if (this.random.nextInt(100) <= chanceToSpawn) {
                         if (this.spawn(level)) {
-                            data.setSpawnChance(25);
+                            data.setSpawnChance(level.paperConfig().entities.spawning.wanderingTrader.spawnChanceMin);
+                            // Paper end - Add Wandering Trader spawn rate config options
                         }
                     }
                 }
-            }
+            } else { this.tickDelay--; } // Paper - Prevent tickDelay going below 0
         }
     }
 
@@ -93,13 +_,13 @@
                 return false;
             }
 
-            WanderingTrader trader = EntityTypes.WANDERING_TRADER.spawn(level, spawnPosition, EntitySpawnReason.EVENT);
+                WanderingTrader trader = EntityTypes.WANDERING_TRADER.spawn(level, wanderingTrader -> wanderingTrader.setDespawnDelay(48000), spawnPosition, EntitySpawnReason.EVENT, false, false, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit // Paper - set despawnTimer before spawn events called
             if (trader != null) {
                 for (int i = 0; i < 2; i++) {
                     this.tryToSpawnLlamaFor(level, trader, 4);
                 }
 
-                trader.setDespawnDelay(48000);
+                    // trader.setDespawnDelay(48000); // Paper - moved above, modifiable by plugins on CreatureSpawnEvent
                 trader.setWanderTarget(referencePos);
                 trader.setHomeTo(referencePos, 16);
                 return true;
@@ -112,7 +_,7 @@
     private void tryToSpawnLlamaFor(ServerLevel level, WanderingTrader trader, int radius) {
         BlockPos spawnPosition = this.findSpawnPositionNear(level, trader.blockPosition(), radius);
         if (spawnPosition != null) {
-            TraderLlama llama = EntityTypes.TRADER_LLAMA.spawn(level, spawnPosition, EntitySpawnReason.EVENT);
+            TraderLlama llama = EntityTypes.TRADER_LLAMA.spawn(level, spawnPosition, EntitySpawnReason.EVENT, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL); // CraftBukkit
             if (llama != null) {
                 llama.setLeashedTo(trader, true);
             }
@@ -127,7 +_,17 @@
             int xPosition = referencePosition.getX() + this.random.nextInt(radius * 2) - radius;
             int zPosition = referencePosition.getZ() + this.random.nextInt(radius * 2) - radius;
             int yPosition = level.getHeight(SpawnPlacements.getHeightmapType(EntityTypes.WANDERING_TRADER), xPosition, zPosition);
-            BlockPos spawnPos = new BlockPos(xPosition, yPosition, zPosition);
+            // Purpur start - Allow toggling special MobSpawners per world - allow traders to spawn below nether roof
+            BlockPos.MutableBlockPos spawnPos = new BlockPos.MutableBlockPos(xPosition, yPosition, zPosition);
+            if (level.dimensionType().hasCeiling()) {
+                do {
+                    spawnPos.relative(net.minecraft.core.Direction.DOWN);
+                } while (!level.getBlockState(spawnPos).isAir());
+                do {
+                    spawnPos.relative(net.minecraft.core.Direction.DOWN);
+                } while (level.getBlockState(spawnPos).isAir() && spawnPos.getY() > 0);
+            }
+            // Purpur end - Allow toggling special MobSpawners per world
             if (wanderingTraderSpawnType.isSpawnPositionOk(level, spawnPos, EntityTypes.WANDERING_TRADER)) {
                 spawnPosition = spawnPos;
                 break;
@@ -137,8 +_,13 @@
         return spawnPosition;
     }
 
-    private boolean hasEnoughSpace(LevelReader level, BlockPos spawnPos) {
-        return !level.findBlocksIn(spawnPos, spawnPos.offset(1, 2, 1))
-            .forEachUntil((pos, state) -> Continuation.continueIf(state.getCollisionShape(level, pos).isEmpty()));
+    private boolean hasEnoughSpace(BlockGetter level, BlockPos spawnPos) {
+        for (BlockPos pos : BlockPos.betweenClosed(spawnPos, spawnPos.offset(1, 2, 1))) {
+            if (!level.getBlockState(pos).getCollisionShape(level, pos).isEmpty()) {
+                return false;
+            }
+        }
+
+        return true;
     }
 }
