--- a/net/minecraft/world/entity/monster/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/AbstractSkeleton.java
@@ -42,6 +_,7 @@
 import net.minecraft.world.level.ServerLevelAccessor;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public abstract class AbstractSkeleton extends Monster implements RangedAttackMob {
     private static final int HARD_ATTACK_INTERVAL = 20;
@@ -64,6 +_,7 @@
     protected AbstractSkeleton(EntityType<? extends AbstractSkeleton> p_32133_, Level p_32134_) {
         super(p_32133_, p_32134_);
         this.reassessWeaponGoal();
+        this.setShouldBurnInDay(true); // Purpur - API for any mob to burn daylight
     }
 
     @Override
@@ -91,29 +_,13 @@
 
     protected abstract SoundEvent getStepSound();
 
+    // Paper start - shouldBurnInDay API
+    public boolean shouldBurnInDay() { return shouldBurnInDay; }
+    public void setShouldBurnInDay(boolean shouldBurnInDay) { this.shouldBurnInDay = shouldBurnInDay; }
+    // Paper end - shouldBurnInDay API
+
     @Override
     public void aiStep() {
-        boolean flag = this.isSunBurnTick();
-        if (flag) {
-            ItemStack itemstack = this.getItemBySlot(EquipmentSlot.HEAD);
-            if (!itemstack.isEmpty()) {
-                if (itemstack.isDamageableItem()) {
-                    Item item = itemstack.getItem();
-                    itemstack.setDamageValue(itemstack.getDamageValue() + this.random.nextInt(2));
-                    if (itemstack.getDamageValue() >= itemstack.getMaxDamage()) {
-                        this.onEquippedItemBroken(item, EquipmentSlot.HEAD);
-                        this.setItemSlot(EquipmentSlot.HEAD, ItemStack.EMPTY);
-                    }
-                }
-
-                flag = false;
-            }
-
-            if (flag) {
-                this.igniteForSeconds(8.0F);
-            }
-        }
-
         super.aiStep();
     }
 
@@ -139,7 +_,7 @@
         this.populateDefaultEquipmentSlots(randomsource, p_32147_);
         this.populateDefaultEquipmentEnchantments(p_32146_, randomsource, p_32147_);
         this.reassessWeaponGoal();
-        this.setCanPickUpLoot(randomsource.nextFloat() < 0.55F * p_32147_.getSpecialMultiplier());
+        this.setCanPickUpLoot(this.level().paperConfig().entities.behavior.mobsCanAlwaysPickUpLoot.skeletons || randomsource.nextFloat() < 0.55F * p_32147_.getSpecialMultiplier());
         if (this.getItemBySlot(EquipmentSlot.HEAD).isEmpty()) {
             LocalDate localdate = LocalDate.now();
             int i = localdate.get(ChronoField.DAY_OF_MONTH);
@@ -157,8 +_,8 @@
         if (this.level() != null && !this.level().isClientSide) {
             this.goalSelector.removeGoal(this.meleeGoal);
             this.goalSelector.removeGoal(this.bowGoal);
-            ItemStack itemstack = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, Items.BOW));
-            if (itemstack.is(Items.BOW)) {
+            ItemStack itemstack = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.world.item.BowItem));
+            if (itemstack.getItem() instanceof net.minecraft.world.item.BowItem) {
                 int i = this.getHardAttackInterval();
                 if (this.level().getDifficulty() != Difficulty.HARD) {
                     i = this.getAttackInterval();
@@ -182,16 +_,29 @@
 
     @Override
     public void performRangedAttack(LivingEntity p_32141_, float p_32142_) {
-        ItemStack itemstack = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, Items.BOW));
-        ItemStack itemstack1 = this.getProjectile(itemstack);
-        AbstractArrow abstractarrow = this.getArrow(itemstack1, p_32142_, itemstack);
+        ItemStack weapon = this.getItemInHand(ProjectileUtil.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.world.item.BowItem));
+        ItemStack itemstack1 = this.getProjectile(weapon);
+        AbstractArrow abstractarrow = this.getArrow(itemstack1, p_32142_, weapon);
+        if (weapon.getItem() instanceof net.minecraft.world.item.ProjectileWeaponItem weaponItem)
+            abstractarrow = weaponItem.customArrow(abstractarrow, itemstack1, weapon);
         double d0 = p_32141_.getX() - this.getX();
         double d1 = p_32141_.getY(0.3333333333333333) - abstractarrow.getY();
         double d2 = p_32141_.getZ() - this.getZ();
         double d3 = Math.sqrt(d0 * d0 + d2 * d2);
         abstractarrow.shoot(d0, d1 + d3 * 0.2F, d2, 1.6F, (float)(14 - this.level().getDifficulty().getId() * 4));
+        // CraftBukkit start
+        org.bukkit.event.entity.EntityShootBowEvent event = CraftEventFactory.callEntityShootBowEvent(this, this.getMainHandItem(), abstractarrow.getPickupItem(), abstractarrow, net.minecraft.world.InteractionHand.MAIN_HAND, 0.8F, true);  // Paper
+        if (event.isCancelled()) {
+            event.getProjectile().remove();
+            return;
+        }
+
+        if (event.getProjectile() == abstractarrow.getBukkitEntity()) {
+            this.level().addFreshEntity(abstractarrow);
+        }
+        // CraftBukkit end
         this.playSound(SoundEvents.SKELETON_SHOOT, 1.0F, 1.0F / (this.getRandom().nextFloat() * 0.4F + 0.8F));
-        this.level().addFreshEntity(abstractarrow);
+        // this.level().addFreshEntity(abstractarrow); // CraftBukkit - moved up
     }
 
     protected AbstractArrow getArrow(ItemStack p_32156_, float p_32157_, @Nullable ItemStack p_346155_) {
@@ -208,6 +_,13 @@
         super.readAdditionalSaveData(p_32152_);
         this.reassessWeaponGoal();
     }
+
+    // Paper start - shouldBurnInDay API
+    @Override
+    public void addAdditionalSaveData(CompoundTag nbt) {
+        super.addAdditionalSaveData(nbt);
+    }
+    // Paper end - shouldBurnInDay API
 
     @Override
     public void setItemSlot(EquipmentSlot p_32138_, ItemStack p_32139_) {
