--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.monster;
 
+import com.destroystokyo.paper.event.entity.CreeperIgniteEvent;
 import java.util.Collection;
 import javax.annotation.Nullable;
 import net.minecraft.nbt.CompoundTag;
@@ -40,6 +_,7 @@
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class Creeper extends Monster implements PowerableMob {
     private static final EntityDataAccessor<Integer> DATA_SWELL_DIR = SynchedEntityData.defineId(Creeper.class, EntityDataSerializers.INT);
@@ -50,6 +_,7 @@
     public int maxSwell = 30;
     public int explosionRadius = 3;
     private int droppedSkulls;
+    public Entity entityIgniter; // CraftBukkit
 
     public Creeper(EntityType<? extends Creeper> p_32278_, Level p_32279_) {
         super(p_32278_, p_32279_);
@@ -122,7 +_,7 @@
         }
 
         if (p_32296_.getBoolean("ignited")) {
-            this.ignite();
+            this.entityData.set(Creeper.DATA_IS_IGNITED, true); // Paper - set directly to avoid firing event
         }
     }
 
@@ -167,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.CREEPER_DEATH;
     }
 
@@ -206,9 +_,20 @@
     @Override
     public void thunderHit(ServerLevel p_32286_, LightningBolt p_32287_) {
         super.thunderHit(p_32286_, p_32287_);
+        // CraftBukkit start
+        if (CraftEventFactory.callCreeperPowerEvent(this, p_32287_, org.bukkit.event.entity.CreeperPowerEvent.PowerCause.LIGHTNING).isCancelled()) {
+            return;
+        }
+        // CraftBukkit end
         this.entityData.set(DATA_IS_POWERED, true);
     }
 
+    // CraftBukkit start
+    public void setPowered(boolean powered) {
+        this.entityData.set(Creeper.DATA_IS_POWERED, powered);
+    }
+    // CraftBukkit end
+
     @Override
     protected InteractionResult mobInteract(Player p_32301_, InteractionHand p_32302_) {
         ItemStack itemstack = p_32301_.getItemInHand(p_32302_);
@@ -217,8 +_,9 @@
             this.level()
                 .playSound(p_32301_, this.getX(), this.getY(), this.getZ(), soundevent, this.getSoundSource(), 1.0F, this.random.nextFloat() * 0.4F + 0.8F);
             if (!this.level().isClientSide) {
+                this.entityIgniter = p_32301_; // CraftBukkit
                 this.ignite();
-                if (!itemstack.isDamageableItem()) {
+                if (itemstack.getMaxDamage() == 0) { // CraftBukkit - fix MC-264285: unbreakable flint and steels are completely consumed when igniting a creeper
                     itemstack.shrink(1);
                 } else {
                     itemstack.hurtAndBreak(1, p_32301_, getSlotForHand(p_32302_));
@@ -234,18 +_,35 @@
     public void explodeCreeper() {
         if (!this.level().isClientSide) {
             float f = this.isPowered() ? 2.0F : 1.0F;
-            this.dead = true;
-            this.level().explode(this, this.getX(), this.getY(), this.getZ(), (float)this.explosionRadius * f, Level.ExplosionInteraction.MOB);
-            this.spawnLingeringCloud();
-            this.triggerOnDeathMobEffects(Entity.RemovalReason.KILLED);
-            this.discard();
+            // CraftBukkit start
+            org.bukkit.event.entity.ExplosionPrimeEvent event = CraftEventFactory.callExplosionPrimeEvent(this, this.explosionRadius * f, false);
+            if (!event.isCancelled()) {
+                // CraftBukkit end
+                this.dead = true;
+                // Youer - fix mixin
+                if (event.getFire()) {
+                    this.level().explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.MOB); // CraftBukkit // Paper - fix DamageSource API (revert to vanilla, no, just no, don't change this)
+                } else {
+                    this.level().explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), Level.ExplosionInteraction.MOB);
+                }
+                // Youer end
+                this.spawnLingeringCloud();
+                this.triggerOnDeathMobEffects(Entity.RemovalReason.KILLED);
+                this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit - add Bukkit remove cause
+                // CraftBukkit start
+            } else {
+                this.swell = 0;
+                this.entityData.set(DATA_IS_IGNITED, Boolean.valueOf(false)); // Paper
+            }
+            // CraftBukkit end
         }
     }
 
     private void spawnLingeringCloud() {
         Collection<MobEffectInstance> collection = this.getActiveEffects();
-        if (!collection.isEmpty()) {
+        if (!collection.isEmpty() && !this.level().paperConfig().entities.behavior.disableCreeperLingeringEffect) { // Paper - Option to disable creeper lingering effect
             AreaEffectCloud areaeffectcloud = new AreaEffectCloud(this.level(), this.getX(), this.getY(), this.getZ());
+            areaeffectcloud.setOwner(this); // CraftBukkit
             areaeffectcloud.setRadius(2.5F);
             areaeffectcloud.setRadiusOnUse(-0.5F);
             areaeffectcloud.setWaitTime(10);
@@ -256,7 +_,7 @@
                 areaeffectcloud.addEffect(new MobEffectInstance(mobeffectinstance));
             }
 
-            this.level().addFreshEntity(areaeffectcloud);
+            this.level().addFreshEntity(areaeffectcloud, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.EXPLOSION); // CraftBukkit
         }
     }
 
@@ -265,7 +_,18 @@
     }
 
     public void ignite() {
-        this.entityData.set(DATA_IS_IGNITED, true);
+        // Paper start - CreeperIgniteEvent
+        setIgnited(true);
+    }
+
+    public void setIgnited(boolean ignited) {
+        if (isIgnited() != ignited) {
+            CreeperIgniteEvent event = new CreeperIgniteEvent((org.bukkit.entity.Creeper) getBukkitEntity(), ignited);
+            if (event.callEvent()) {
+                this.entityData.set(Creeper.DATA_IS_IGNITED, event.isIgnited());
+            }
+        }
+        // Paper end - CreeperIgniteEvent
     }
 
     public boolean canDropMobsSkull() {
