--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -46,9 +_,9 @@
     private static final EntityDataAccessor<Boolean> DATA_IS_POWERED = SynchedEntityData.defineId(Creeper.class, EntityDataSerializers.BOOLEAN);
     private static final EntityDataAccessor<Boolean> DATA_IS_IGNITED = SynchedEntityData.defineId(Creeper.class, EntityDataSerializers.BOOLEAN);
     private int oldSwell;
-    public int swell;
-    public int maxSwell = 30;
-    public int explosionRadius = 3;
+    private int swell;
+    private int maxSwell = 30;
+    private int explosionRadius = 3;
     private int droppedSkulls;
 
     public Creeper(EntityType<? extends Creeper> p_32278_, Level p_32279_) {
@@ -231,7 +_,7 @@
         }
     }
 
-    public void explodeCreeper() {
+    private void explodeCreeper() {
         if (!this.level().isClientSide) {
             float f = this.isPowered() ? 2.0F : 1.0F;
             this.dead = true;
