--- a/net/minecraft/world/entity/monster/Creeper.java
+++ b/net/minecraft/world/entity/monster/Creeper.java
@@ -50,6 +_,7 @@
     public int maxSwell = 30;
     public int explosionRadius = 3;
     private int droppedSkulls;
+    public Entity entityIgniter; // CraftBukkit
 
     public Creeper(EntityType<? extends Creeper> p_32278_, Level p_32279_) {
         super(p_32278_, p_32279_);
@@ -208,6 +_,12 @@
         super.thunderHit(p_32286_, p_32287_);
         this.entityData.set(DATA_IS_POWERED, true);
     }
+
+    // CraftBukkit start
+    public void setPowered(boolean powered) {
+        this.entityData.set(DATA_IS_POWERED, powered);
+    }
+    // CraftBukkit end
 
     @Override
     protected InteractionResult mobInteract(Player p_32301_, InteractionHand p_32302_) {
