--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -54,7 +_,6 @@
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
 import net.minecraft.world.item.enchantment.providers.VanillaEnchantmentProviders;
 import net.minecraft.world.level.ClipContext;
-import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
@@ -66,6 +_,7 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EnderMan extends Monster implements NeutralMob {
     private static final ResourceLocation SPEED_MODIFIER_ATTACKING_ID = ResourceLocation.withDefaultNamespace("attacking");
@@ -118,7 +_,6 @@
 
     @Override
     public void setTarget(@Nullable LivingEntity p_32537_) {
-        super.setTarget(p_32537_);
         AttributeInstance attributeinstance = this.getAttribute(Attributes.MOVEMENT_SPEED);
         if (p_32537_ == null) {
             this.targetChangeTime = 0;
@@ -132,7 +_,14 @@
                 attributeinstance.addTransientModifier(SPEED_MODIFIER_ATTACKING);
             }
         }
-    }
+        super.setTarget(p_32537_); //Forge: Moved down to allow event handlers to write data manager values.
+    }
+
+    // Paper start - EndermanEscapeEvent
+    private boolean tryEscape(com.destroystokyo.paper.event.entity.EndermanEscapeEvent.Reason reason) {
+        return new com.destroystokyo.paper.event.entity.EndermanEscapeEvent((org.bukkit.craftbukkit.entity.CraftEnderman) this.getBukkitEntity(), reason).callEvent();
+    }
+    // Paper end - EndermanEscapeEvent
 
     @Override
     protected void defineSynchedData(SynchedEntityData.Builder p_326272_) {
@@ -213,8 +_,15 @@
     }
 
     boolean isLookingAtMe(Player p_32535_) {
+        boolean shouldAttack = isLookingAtMe_check(p_32535_);
+        com.destroystokyo.paper.event.entity.EndermanAttackPlayerEvent event = new com.destroystokyo.paper.event.entity.EndermanAttackPlayerEvent((org.bukkit.entity.Enderman) getBukkitEntity(), (org.bukkit.entity.Player) p_32535_.getBukkitEntity());
+        event.setCancelled(!shouldAttack);
+        return event.callEvent();
+    }
+    private boolean isLookingAtMe_check(Player p_32535_) {
+        // Paper end - EndermanAttackPlayerEvent
         ItemStack itemstack = p_32535_.getInventory().armor.get(3);
-        if (itemstack.is(Blocks.CARVED_PUMPKIN.asItem())) {
+        if (net.neoforged.neoforge.common.CommonHooks.shouldSuppressEnderManAnger(this, p_32535_, itemstack)) {
             return false;
         } else {
             Vec3 vec3 = p_32535_.getViewVector(1.0F).normalize();
@@ -260,7 +_,7 @@
     protected void customServerAiStep() {
         if (this.level().isDay() && this.tickCount >= this.targetChangeTime + 600) {
             float f = this.getLightLevelDependentMagicValue();
-            if (f > 0.5F && this.level().canSeeSky(this.blockPosition()) && this.random.nextFloat() * 30.0F < (f - 0.4F) * 2.0F) {
+            if (f > 0.5F && this.level().canSeeSky(this.blockPosition()) && this.random.nextFloat() * 30.0F < (f - 0.4F) * 2.0F && this.tryEscape(com.destroystokyo.paper.event.entity.EndermanEscapeEvent.Reason.RUNAWAY)) { // Paper - EndermanEscapeEvent
                 this.setTarget(null);
                 this.teleport();
             }
@@ -301,8 +_,10 @@
         boolean flag = blockstate.blocksMotion();
         boolean flag1 = blockstate.getFluidState().is(FluidTags.WATER);
         if (flag && !flag1) {
+            net.neoforged.neoforge.event.entity.EntityTeleportEvent.EnderEntity event = net.neoforged.neoforge.event.EventHooks.onEnderTeleport(this, p_32544_, p_32545_, p_32546_);
+            if (event.isCanceled()) return false;
             Vec3 vec3 = this.position();
-            boolean flag2 = this.randomTeleport(p_32544_, p_32545_, p_32546_, true);
+            boolean flag2 = this.randomTeleport(event.getTargetX(), event.getTargetY(), event.getTargetZ(), true);
             if (flag2) {
                 this.level().gameEvent(GameEvent.TELEPORT, vec3, GameEvent.Context.of(this));
                 if (!this.isSilent()) {
@@ -328,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.ENDERMAN_DEATH;
     }
 
@@ -381,11 +_,13 @@
             } else {
                 boolean flag1 = flag && this.hurtWithCleanWater(p_32494_, (ThrownPotion)p_32494_.getDirectEntity(), p_32495_);
 
-                for (int i = 0; i < 64; i++) {
+                if (this.tryEscape(com.destroystokyo.paper.event.entity.EndermanEscapeEvent.Reason.INDIRECT)) { // Paper - EndermanEscapeEvent
+                    for (int i = 0; i < 64; i++) {
                     if (this.teleport()) {
                         return true;
                     }
                 }
+                } // Paper - EndermanEscapeEvent
 
                 return flag1;
             }
@@ -410,6 +_,16 @@
         this.entityData.set(DATA_STARED_AT, true);
     }
 
+    // Paper start
+    public void setCreepy(boolean creepy) {
+        this.entityData.set(EnderMan.DATA_CREEPY, creepy);
+    }
+
+    public void setHasBeenStaredAt(boolean hasBeenStaredAt) {
+        this.entityData.set(EnderMan.DATA_STARED_AT, hasBeenStaredAt);
+    }
+    // Paper end
+
     @Override
     public boolean requiresCustomPersistence() {
         return super.requiresCustomPersistence() || this.getCarriedBlock() != null;
@@ -459,7 +_,7 @@
             if (this.enderman.getCarriedBlock() == null) {
                 return false;
             } else {
-                return !this.enderman.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+                return !net.neoforged.neoforge.event.EventHooks.canEntityGrief(this.enderman.level(), this.enderman)
                     ? false
                     : this.enderman.getRandom().nextInt(reducedTickDelay(2000)) == 0;
             }
@@ -479,10 +_,12 @@
             BlockState blockstate2 = this.enderman.getCarriedBlock();
             if (blockstate2 != null) {
                 blockstate2 = Block.updateFromNeighbourShapes(blockstate2, this.enderman.level(), blockpos);
-                if (this.canPlaceBlock(level, blockpos, blockstate2, blockstate, blockstate1, blockpos1)) {
-                    level.setBlock(blockpos, blockstate2, 3);
-                    level.gameEvent(GameEvent.BLOCK_PLACE, blockpos, GameEvent.Context.of(this.enderman, blockstate2));
-                    this.enderman.setCarriedBlock(null);
+                if (this.canPlaceBlock(level, blockpos, blockstate2, blockstate, blockstate1, blockpos1) && !net.neoforged.neoforge.event.EventHooks.onBlockPlace(enderman, net.neoforged.neoforge.common.util.BlockSnapshot.create(level.dimension(), level, blockpos1), net.minecraft.core.Direction.UP)) {
+                    if (CraftEventFactory.callEntityChangeBlockEvent(this.enderman, blockpos, blockstate2)) { // CraftBukkit - Place event
+                        level.setBlock(blockpos, blockstate2, 3);
+                        level.gameEvent(GameEvent.BLOCK_PLACE, blockpos, GameEvent.Context.of(this.enderman, blockstate2));
+                        this.enderman.setCarriedBlock(null);
+                    } // CraftBukkit
                 }
             }
         }
@@ -491,6 +_,7 @@
             return p_32562_.isAir()
                 && !p_32563_.isAir()
                 && !p_32563_.is(Blocks.BEDROCK)
+                && !p_32563_.is(net.neoforged.neoforge.common.Tags.Blocks.ENDERMAN_PLACE_ON_BLACKLIST)
                 && p_32563_.isCollisionShapeFullBlock(p_32559_, p_32564_)
                 && p_32561_.canSurvive(p_32559_, p_32560_)
                 && p_32559_.getEntities(this.enderman, AABB.unitCubeFromLowerCorner(Vec3.atLowerCornerOf(p_32560_))).isEmpty();
@@ -573,7 +_,7 @@
             } else {
                 if (this.target != null && !this.enderman.isPassenger()) {
                     if (this.enderman.isLookingAtMe((Player)this.target)) {
-                        if (this.target.distanceToSqr(this.enderman) < 16.0) {
+                        if (this.target.distanceToSqr(this.enderman) < 16.0 && this.enderman.tryEscape(com.destroystokyo.paper.event.entity.EndermanEscapeEvent.Reason.STARE)) { // Paper - EndermanEscapeEvent
                             this.enderman.teleport();
                         }
 
@@ -602,7 +_,7 @@
             if (this.enderman.getCarriedBlock() != null) {
                 return false;
             } else {
-                return !this.enderman.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)
+                return !net.neoforged.neoforge.event.EventHooks.canEntityGrief(this.enderman.level(), this.enderman)
                     ? false
                     : this.enderman.getRandom().nextInt(reducedTickDelay(20)) == 0;
             }
@@ -622,9 +_,11 @@
             BlockHitResult blockhitresult = level.clip(new ClipContext(vec3, vec31, ClipContext.Block.OUTLINE, ClipContext.Fluid.NONE, this.enderman));
             boolean flag = blockhitresult.getBlockPos().equals(blockpos);
             if (blockstate.is(BlockTags.ENDERMAN_HOLDABLE) && flag) {
-                level.removeBlock(blockpos, false);
-                level.gameEvent(GameEvent.BLOCK_DESTROY, blockpos, GameEvent.Context.of(this.enderman, blockstate));
-                this.enderman.setCarriedBlock(blockstate.getBlock().defaultBlockState());
+                if (CraftEventFactory.callEntityChangeBlockEvent(this.enderman, blockpos, blockstate.getFluidState().createLegacyBlock())) { // CraftBukkit - Place event // Paper - fix wrong block state
+                    level.removeBlock(blockpos, false);
+                    level.gameEvent(GameEvent.BLOCK_DESTROY, blockpos, GameEvent.Context.of(this.enderman, blockstate));
+                    this.enderman.setCarriedBlock(blockstate.getBlock().defaultBlockState());
+                } // CraftBukkit
             }
         }
     }
