--- a/net/minecraft/world/entity/monster/Evoker.java
+++ b/net/minecraft/world/entity/monster/Evoker.java
@@ -115,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.EVOKER_DEATH;
     }
 
@@ -124,12 +_,12 @@
         return SoundEvents.EVOKER_HURT;
     }
 
-    void setWololoTarget(@Nullable Sheep p_32635_) {
+    public void setWololoTarget(@Nullable Sheep p_32635_) {
         this.wololoTarget = p_32635_;
     }
 
     @Nullable
-    Sheep getWololoTarget() {
+    public Sheep getWololoTarget() {
         return this.wololoTarget;
     }
 
@@ -275,7 +_,7 @@
                         serverlevel.getScoreboard().addPlayerToTeam(vex.getScoreboardName(), playerteam);
                     }
 
-                    serverlevel.addFreshEntityWithPassengers(vex);
+                    serverlevel.addFreshEntityWithPassengers(vex, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SPELL); // CraftBukkit - Add SpawnReason
                     serverlevel.gameEvent(GameEvent.ENTITY_PLACE, blockpos, GameEvent.Context.of(Evoker.this));
                 }
             }
@@ -305,7 +_,7 @@
                 return false;
             } else if (Evoker.this.tickCount < this.nextAttackTickCount) {
                 return false;
-            } else if (!Evoker.this.level().getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            } else if (!net.neoforged.neoforge.event.EventHooks.canEntityGrief(Evoker.this.level(), Evoker.this)) {
                 return false;
             } else {
                 List<Sheep> list = Evoker.this.level()
