--- a/net/minecraft/world/entity/monster/Ghast.java
+++ b/net/minecraft/world/entity/monster/Ghast.java
@@ -132,22 +_,22 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.GHAST_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_32750_) {
+    public SoundEvent getHurtSound(DamageSource p_32750_) {
         return SoundEvents.GHAST_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.GHAST_DEATH;
     }
 
     @Override
-    protected float getSoundVolume() {
+    public float getSoundVolume() {
         return 5.0F;
     }
 
