--- a/net/minecraft/world/entity/monster/Ghast.java
+++ b/net/minecraft/world/entity/monster/Ghast.java
@@ -73,6 +_,12 @@
         return this.explosionPower;
     }
 
+    // Paper start
+    public void setExplosionPower(int explosionPower) {
+        this.explosionPower = explosionPower;
+    }
+    // Paper end
+
     private static boolean isReflectedFireball(DamageSource source) {
         return source.getDirectEntity() instanceof LargeFireball && source.getEntity() instanceof Player;
     }
@@ -147,6 +_,11 @@
     }
 
     public static boolean checkGhastSpawnRules(EntityType<Ghast> type, LevelAccessor level, EntitySpawnReason spawnReason, BlockPos pos, RandomSource random) {
+        // Purpur start - Config to disable hostile mob spawn on ice
+        if (net.minecraft.world.entity.monster.Monster.canSpawnInBlueAndPackedIce(level, pos)) {
+            return false;
+        }
+        // Purpur end - Config to disable hostile mob spawn on ice
         return level.getDifficulty() != Difficulty.PEACEFUL && random.nextInt(20) == 0 && checkMobSpawnRules(type, level, spawnReason, pos, random);
     }
 
@@ -367,6 +_,8 @@
                         }
 
                         LargeFireball entity = new LargeFireball(level, this.ghast, direction.normalize(), this.ghast.getExplosionPower());
+                        // CraftBukkit - set bukkitYield when setting explosionpower
+                        entity.bukkitYield = entity.explosionPower = this.ghast.getExplosionPower();
                         entity.setPos(this.ghast.getX() + viewVector.x * 4.0, this.ghast.getY(0.5) + 0.5, entity.getZ() + viewVector.z * 4.0);
                         level.addFreshEntity(entity);
                         this.chargeTime = -40;
