--- a/net/minecraft/world/entity/monster/Guardian.java
+++ b/net/minecraft/world/entity/monster/Guardian.java
@@ -57,7 +_,7 @@
     private int clientSideAttackTime;
     private boolean clientSideTouchedGround;
     @Nullable
-    public RandomStrollGoal randomStrollGoal;
+    protected RandomStrollGoal randomStrollGoal;
 
     public Guardian(EntityType<? extends Guardian> p_32810_, Level p_32811_) {
         super(p_32810_, p_32811_);
@@ -115,7 +_,7 @@
         return 80;
     }
 
-    public void setActiveAttackTarget(int p_32818_) {
+    void setActiveAttackTarget(int p_32818_) {
         this.entityData.set(DATA_ID_ATTACK_TARGET, p_32818_);
     }
 
@@ -356,9 +_,9 @@
         }
     }
 
-    public static class GuardianAttackGoal extends Goal {
+    static class GuardianAttackGoal extends Goal {
         private final Guardian guardian;
-        public int attackTime;
+        private int attackTime;
         private final boolean elder;
 
         public GuardianAttackGoal(Guardian p_32871_) {
