--- a/net/minecraft/world/entity/monster/Husk.java
+++ b/net/minecraft/world/entity/monster/Husk.java
@@ -28,22 +_,22 @@
     }
 
     @Override
-    protected boolean isSunSensitive() {
+    public boolean isSunSensitive() {
         return false;
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.HUSK_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_32903_) {
+    public SoundEvent getHurtSound(DamageSource p_32903_) {
         return SoundEvents.HUSK_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.HUSK_DEATH;
     }
 
@@ -70,6 +_,7 @@
 
     @Override
     protected void doUnderWaterConversion() {
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.ZOMBIE, (timer) -> this.conversionTime = timer)) return;
         this.convertToZombieType(EntityType.ZOMBIE);
         if (!this.isSilent()) {
             this.level().levelEvent(null, 1041, this.blockPosition(), 0);
