--- a/net/minecraft/world/entity/monster/Phantom.java
+++ b/net/minecraft/world/entity/monster/Phantom.java
@@ -243,6 +_,7 @@
 
                     for (Player player : list) {
                         if (Phantom.this.canAttack(player, TargetingConditions.DEFAULT)) {
+                            Phantom.this.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER, true); // CraftBukkit - reason
                             Phantom.this.setTarget(player);
                             return true;
                         }
