--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -61,7 +_,7 @@
     );
     protected static final EntityDataAccessor<Direction> DATA_ATTACH_FACE_ID = SynchedEntityData.defineId(Shulker.class, EntityDataSerializers.DIRECTION);
     protected static final EntityDataAccessor<Byte> DATA_PEEK_ID = SynchedEntityData.defineId(Shulker.class, EntityDataSerializers.BYTE);
-    public static final EntityDataAccessor<Byte> DATA_COLOR_ID = SynchedEntityData.defineId(Shulker.class, EntityDataSerializers.BYTE);
+    protected static final EntityDataAccessor<Byte> DATA_COLOR_ID = SynchedEntityData.defineId(Shulker.class, EntityDataSerializers.BYTE);
     private static final int TELEPORT_STEPS = 6;
     private static final byte NO_COLOR = 16;
     private static final byte DEFAULT_COLOR = 16;
@@ -392,6 +_,12 @@
                     && this.level().noCollision(this, new AABB(blockpos1).deflate(1.0E-6))) {
                     Direction direction = this.findAttachableSurface(blockpos1);
                     if (direction != null) {
+                        net.neoforged.neoforge.event.entity.EntityTeleportEvent.EnderEntity event = net.neoforged.neoforge.event.EventHooks.onEnderTeleport(this, blockpos1.getX(), blockpos1.getY(), blockpos1.getZ());
+                        if (event.isCanceled()) direction = null;
+                        blockpos1 = BlockPos.containing(event.getTargetX(), event.getTargetY(), event.getTargetZ());
+                    }
+
+                    if (direction != null) {
                         this.unRide();
                         this.setAttachFace(direction);
                         this.playSound(SoundEvents.SHULKER_TELEPORT, 1.0F, 1.0F);
@@ -472,7 +_,7 @@
         return this.entityData.get(DATA_ATTACH_FACE_ID);
     }
 
-    public void setAttachFace(Direction p_149789_) {
+    private void setAttachFace(Direction p_149789_) {
         this.entityData.set(DATA_ATTACH_FACE_ID, p_149789_);
     }
 
@@ -485,11 +_,11 @@
         super.onSyncedDataUpdated(p_33434_);
     }
 
-    public int getRawPeekAmount() {
+    private int getRawPeekAmount() {
         return this.entityData.get(DATA_PEEK_ID);
     }
 
-    public void setRawPeekAmount(int p_33419_) {
+    void setRawPeekAmount(int p_33419_) {
         if (!this.level().isClientSide) {
             this.getAttribute(Attributes.ARMOR).removeModifier(COVERED_ARMOR_MODIFIER_ID);
             if (p_33419_ == 0) {
