--- a/net/minecraft/world/entity/monster/Skeleton.java
+++ b/net/minecraft/world/entity/monster/Skeleton.java
@@ -14,10 +_,10 @@
 
 public class Skeleton extends AbstractSkeleton {
     private static final int TOTAL_CONVERSION_TIME = 300;
-    public static final EntityDataAccessor<Boolean> DATA_STRAY_CONVERSION_ID = SynchedEntityData.defineId(Skeleton.class, EntityDataSerializers.BOOLEAN);
+    private static final EntityDataAccessor<Boolean> DATA_STRAY_CONVERSION_ID = SynchedEntityData.defineId(Skeleton.class, EntityDataSerializers.BOOLEAN);
     public static final String CONVERSION_TAG = "StrayConversionTime";
     private int inPowderSnowTime;
-    public int conversionTime;
+    private int conversionTime;
 
     public Skeleton(EntityType<? extends Skeleton> p_33570_, Level p_33571_) {
         super(p_33570_, p_33571_);
@@ -80,13 +_,15 @@
         }
     }
 
-    public void startFreezeConversion(int p_149831_) {
+    private void startFreezeConversion(int p_149831_) {
         this.conversionTime = p_149831_;
         this.setFreezeConverting(true);
     }
 
     protected void doFreezeConversion() {
-        this.convertTo(EntityType.STRAY, true);
+        if (!net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.STRAY, (timer) -> this.conversionTime = timer)) return;
+        Stray stray = this.convertTo(EntityType.STRAY, true);
+        if (stray != null) net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, stray);
         if (!this.isSilent()) {
             this.level().levelEvent(null, 1048, this.blockPosition(), 0);
         }
