--- a/net/minecraft/world/entity/monster/SpellcasterIllager.java
+++ b/net/minecraft/world/entity/monster/SpellcasterIllager.java
@@ -62,7 +_,7 @@
         this.entityData.set(DATA_SPELL_CASTING_ID, (byte)p_33728_.id);
     }
 
-    public SpellcasterIllager.IllagerSpell getCurrentSpell() {
+    protected SpellcasterIllager.IllagerSpell getCurrentSpell() {
         return !this.level().isClientSide ? this.currentSpell : SpellcasterIllager.IllagerSpell.byId(this.entityData.get(DATA_SPELL_CASTING_ID));
     }
 
@@ -116,7 +_,7 @@
 
     protected abstract SoundEvent getCastingSoundEvent();
 
-    public static enum IllagerSpell {
+    protected static enum IllagerSpell {
         NONE(0, 0.0, 0.0, 0.0),
         SUMMON_VEX(1, 0.7, 0.7, 0.8),
         FANGS(2, 0.4, 0.3, 0.35),
@@ -173,8 +_,7 @@
         }
     }
 
-    public abstract class SpellcasterUseSpellGoal extends Goal {
-        protected SpellcasterUseSpellGoal() {}
+    protected abstract class SpellcasterUseSpellGoal extends Goal {
         protected int attackWarmupDelay;
         protected int nextAttackTickCount;
 
