--- a/net/minecraft/world/entity/monster/Spider.java
+++ b/net/minecraft/world/entity/monster/Spider.java
@@ -98,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.SPIDER_DEATH;
     }
 
@@ -121,7 +_,7 @@
 
     @Override
     public boolean canBeAffected(MobEffectInstance p_33809_) {
-        return p_33809_.is(MobEffects.POISON) ? false : super.canBeAffected(p_33809_);
+        return p_33809_.is(MobEffects.POISON) && this.level().paperConfig().entities.mobEffects.spidersImmuneToPoisonEffect ? false : super.canBeAffected(p_33809_); // Paper - Add config for mobs immune to default effects
     }
 
     public boolean isClimbing() {
@@ -163,6 +_,8 @@
         if (p_33793_ instanceof Spider.SpiderEffectsGroupData spider$spidereffectsgroupdata) {
             Holder<MobEffect> holder = spider$spidereffectsgroupdata.effect;
             if (holder != null) {
+                this.addEffectCause(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.SPIDER_SPAWN);
+                this.AddEffect$fireEvent(p_33790_ instanceof net.minecraft.server.level.ServerLevel);
                 this.addEffect(new MobEffectInstance(holder, -1));
             }
         }
