--- a/net/minecraft/world/entity/monster/Stray.java
+++ b/net/minecraft/world/entity/monster/Stray.java
@@ -36,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.STRAY_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_33850_) {
+    public SoundEvent getHurtSound(DamageSource p_33850_) {
         return SoundEvents.STRAY_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.STRAY_DEATH;
     }
 
