--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -70,6 +_,7 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class Strider extends Animal implements ItemSteerable, Saddleable {
     private static final ResourceLocation SUFFOCATING_MODIFIER_ID = ResourceLocation.withDefaultNamespace("suffocating");
@@ -321,7 +_,14 @@
             }
 
             boolean flag1 = flag2;
-            this.setSuffocating(!flag || flag1);
+            // CraftBukkit start
+            boolean suffocating = !flag || flag1;
+            if (suffocating ^ this.isSuffocating()) {
+                if (CraftEventFactory.callStriderTemperatureChangeEvent(this, suffocating)) {
+                    this.setSuffocating(suffocating);
+                }
+            }
+            // CraftBukkit end
         }
 
         super.tick();
@@ -365,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.STRIDER_DEATH;
     }
 
