--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -329,17 +_,17 @@
 
     @Nullable
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return !this.isPanicking() && !this.isBeingTempted() ? SoundEvents.STRIDER_AMBIENT : null;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_33934_) {
+    public SoundEvent getHurtSound(DamageSource p_33934_) {
         return SoundEvents.STRIDER_HURT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.STRIDER_DEATH;
     }
 
