--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -192,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.VEX_DEATH;
     }
 
@@ -296,6 +_,7 @@
 
         @Override
         public void start() {
+            Vex.this.setTargetReason(org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit
             Vex.this.setTarget(Vex.this.owner.getTarget());
             super.start();
         }
