--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -290,7 +_,7 @@
         @Override
         public void start() {
             Mob owner = Vex.this.getOwner();
-            Vex.this.setTarget(owner != null ? owner.getTarget() : null);
+            Vex.this.setTarget(owner != null ? owner.getTarget() : null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit
             super.start();
         }
     }
