--- a/net/minecraft/world/entity/monster/Vindicator.java
+++ b/net/minecraft/world/entity/monster/Vindicator.java
@@ -150,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return SoundEvents.VINDICATOR_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.VINDICATOR_DEATH;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_34103_) {
+    public SoundEvent getHurtSound(DamageSource p_34103_) {
         return SoundEvents.VINDICATOR_HURT;
     }
 
