--- a/net/minecraft/world/entity/monster/Vindicator.java
+++ b/net/minecraft/world/entity/monster/Vindicator.java
@@ -46,7 +_,7 @@
 
 public class Vindicator extends AbstractIllager {
     private static final String TAG_JOHNNY = "Johnny";
-    static final Predicate<Difficulty> DOOR_BREAKING_PREDICATE = p_34082_ -> p_34082_ == Difficulty.NORMAL || p_34082_ == Difficulty.HARD;
+    public static final Predicate<Difficulty> DOOR_BREAKING_PREDICATE = p_34082_ -> p_34082_ == Difficulty.NORMAL || p_34082_ == Difficulty.HARD;
     public boolean isJohnny;
 
     public Vindicator(EntityType<? extends Vindicator> p_34074_, Level p_34075_) {
@@ -151,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.VINDICATOR_DEATH;
     }
 
@@ -179,7 +_,7 @@
 
     static class VindicatorBreakDoorGoal extends BreakDoorGoal {
         public VindicatorBreakDoorGoal(Mob p_34112_) {
-            super(p_34112_, 6, Vindicator.DOOR_BREAKING_PREDICATE);
+            super(p_34112_, 6, com.google.common.base.Predicates.in(p_34112_.level().paperConfig().entities.behavior.doorBreakingDifficulty.getOrDefault(p_34112_.getType(), p_34112_.level().paperConfig().entities.behavior.doorBreakingDifficulty.get(EntityType.VINDICATOR)))); // Paper - Configurable door breaking difficulty
             this.setFlags(EnumSet.of(Goal.Flag.MOVE));
         }
 
