--- a/net/minecraft/world/entity/monster/Witch.java
+++ b/net/minecraft/world/entity/monster/Witch.java
@@ -122,7 +_,7 @@
                     this.setItemSlot(EquipmentSlot.MAINHAND, ItemStack.EMPTY);
                     PotionContents potioncontents = itemstack.get(DataComponents.POTION_CONTENTS);
                     if (itemstack.is(Items.POTION) && potioncontents != null) {
-                        potioncontents.forEachEffect(this::addEffect);
+                        potioncontents.forEachEffect((effect) -> this.addEffect(effect, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK)); // CraftBukkit
                     }
 
                     this.gameEvent(GameEvent.DRINK);
