--- a/net/minecraft/world/entity/monster/WitherSkeleton.java
+++ b/net/minecraft/world/entity/monster/WitherSkeleton.java
@@ -48,7 +_,7 @@
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.WITHER_SKELETON_DEATH;
     }
 
@@ -90,7 +_,7 @@
             return false;
         } else {
             if (p_34169_ instanceof LivingEntity) {
-                ((LivingEntity)p_34169_).addEffect(new MobEffectInstance(MobEffects.WITHER, 200), this);
+                ((LivingEntity)p_34169_).addEffect(new MobEffectInstance(MobEffects.WITHER, 200), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             }
 
             return true;
@@ -106,6 +_,6 @@
 
     @Override
     public boolean canBeAffected(MobEffectInstance p_34192_) {
-        return p_34192_.is(MobEffects.WITHER) ? false : super.canBeAffected(p_34192_);
+        return p_34192_.is(MobEffects.WITHER) && this.level().paperConfig().entities.mobEffects.immuneToWitherEffect.witherSkeleton ? false : super.canBeAffected(p_34192_); // Paper - Add config for mobs immune to default effects
     }
 }
