--- a/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -52,7 +_,7 @@
 
 public class ZombieVillager extends Zombie implements VillagerDataHolder {
     private static final Logger LOGGER = LogUtils.getLogger();
-    public static final EntityDataAccessor<Boolean> DATA_CONVERTING_ID = SynchedEntityData.defineId(ZombieVillager.class, EntityDataSerializers.BOOLEAN);
+    private static final EntityDataAccessor<Boolean> DATA_CONVERTING_ID = SynchedEntityData.defineId(ZombieVillager.class, EntityDataSerializers.BOOLEAN);
     private static final EntityDataAccessor<VillagerData> DATA_VILLAGER_DATA = SynchedEntityData.defineId(
         ZombieVillager.class, EntityDataSerializers.VILLAGER_DATA
     );
@@ -60,9 +_,9 @@
     private static final int VILLAGER_CONVERSION_WAIT_MAX = 6000;
     private static final int MAX_SPECIAL_BLOCKS_COUNT = 14;
     private static final int SPECIAL_BLOCK_RADIUS = 4;
-    public int villagerConversionTime;
+    private int villagerConversionTime;
     @Nullable
-    public UUID conversionStarter;
+    private UUID conversionStarter;
     @Nullable
     private Tag gossips;
     @Nullable
@@ -141,7 +_,7 @@
         if (!this.level().isClientSide && this.isAlive() && this.isConverting()) {
             int i = this.getConversionProgress();
             this.villagerConversionTime -= i;
-            if (this.villagerConversionTime <= 0) {
+            if (this.villagerConversionTime <= 0 && net.neoforged.neoforge.event.EventHooks.canLivingConvert(this, EntityType.VILLAGER, (timer) -> this.villagerConversionTime = timer)) {
                 this.finishConversion((ServerLevel)this.level());
             }
         }
@@ -182,7 +_,7 @@
         return this.getEntityData().get(DATA_CONVERTING_ID);
     }
 
-    public void startConverting(@Nullable UUID p_34384_, int p_34385_) {
+    private void startConverting(@Nullable UUID p_34384_, int p_34385_) {
         this.conversionStarter = p_34384_;
         this.villagerConversionTime = p_34385_;
         this.getEntityData().set(DATA_CONVERTING_ID, true);
@@ -246,6 +_,7 @@
             if (!this.isSilent()) {
                 p_34399_.levelEvent(null, 1027, this.blockPosition(), 0);
             }
+            net.neoforged.neoforge.event.EventHooks.onLivingConvert(this, villager);
         }
     }
 
