--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrowableItemProjectile.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrowableItemProjectile.java
@@ -36,7 +_,13 @@
         this.getEntityData().set(DATA_ITEM_STACK, source.copyWithCount(1));
     }
 
-    protected abstract Item getDefaultItem();
+    public abstract Item getDefaultItem();
+
+    // CraftBukkit start
+    public Item getDefaultItemPublic() {
+        return getDefaultItem();
+    }
+    // CraftBukkit end
 
     @Override
     public ItemStack getItem() {
