--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrowableItemProjectile.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrowableItemProjectile.java
@@ -38,6 +_,12 @@
 
     public abstract Item getDefaultItem();
 
+    // CraftBukkit start
+    public Item getDefaultItemPublic() {
+        return getDefaultItem();
+    }
+    // CraftBukkit end
+
     @Override
     public ItemStack getItem() {
         return this.getEntityData().get(DATA_ITEM_STACK);
