--- a/net/minecraft/network/protocol/Packet.java
+++ b/net/minecraft/network/protocol/Packet.java
@@ -11,6 +_,19 @@
 
     void handle(T listener);
 
+    // Paper start
+    default boolean hasLargePacketFallback() {
+        return false;
+    }
+
+    /**
+     * override {@link #hasLargePacketFallback()} to return true when overriding in subclasses
+     */
+    default boolean packetTooLarge(net.minecraft.network.Connection manager) {
+        return false;
+    }
+    // Paper end
+
     default boolean isSkippable() {
         return false;
     }
@@ -22,4 +_,31 @@
     static <B extends ByteBuf, T extends Packet<?>> StreamCodec<B, T> codec(StreamMemberEncoder<B, T> writer, StreamDecoder<B, T> reader) {
         return StreamCodec.ofMember(writer, reader);
     }
+
+    // Paper start
+    /**
+     * @param player null if not at PLAY stage yet
+     */
+    default void onPacketDispatch(net.minecraft.server.level.@org.jspecify.annotations.Nullable ServerPlayer player) {
+    }
+
+    /**
+     * @param player null if not at PLAY stage yet
+     * @param future can be null if packet was cancelled
+     */
+    default void onPacketDispatchFinish(net.minecraft.server.level.@org.jspecify.annotations.Nullable ServerPlayer player, io.netty.channel.@org.jspecify.annotations.Nullable ChannelFuture future) {
+    }
+
+    default boolean hasFinishListener() {
+        return false;
+    }
+
+    default boolean isReady() {
+        return true;
+    }
+
+    default java.util.@org.jspecify.annotations.Nullable List<Packet<?>> getExtraPackets() {
+        return null;
+    }
+    // Paper end
 }
