--- a/com/mojang/renderpearl/backend/vulkan/DestructionQueue.java
+++ b/com/mojang/renderpearl/backend/vulkan/DestructionQueue.java
@@ -1,13 +_,12 @@
 package com.mojang.renderpearl.backend.vulkan;
 
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import it.unimi.dsi.fastutil.objects.ReferenceArrayList;
 import it.unimi.dsi.fastutil.objects.ReferenceList;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 
 @OnlyIn(Dist.CLIENT)
-public class DestructionQueue<T> implements UncheckedAutoCloseable {
+public class DestructionQueue<T> implements AutoCloseable {
     private final DestructionQueue.Destroyer<T> destroyCallback;
     private final ReferenceList<ReferenceArrayList<T>> destructionQueues;
     private int currentDestructionQueueIndex = 0;
