--- a/com/mojang/renderpearl/backend/vulkan/VulkanGpuTexture.java
+++ b/com/mojang/renderpearl/backend/vulkan/VulkanGpuTexture.java
@@ -68,7 +_,7 @@
             subresourceRange.levelCount(this.getMipLevels());
             subresourceRange.baseArrayLayer(0);
             subresourceRange.layerCount(depthOrLayers);
-            VK12.vkCmdPipelineBarrier(device.createCommandEncoder().objectInitCommandBuffer(), 1, 8192, 0, null, null, barrier);
+            VK12.vkCmdPipelineBarrier(device.createCommandEncoder().textureInitCommandBuffer(), 1, 65536, 0, null, null, barrier);
             device.instance().debug().setObjectName(device.vkDevice(), 10, this.vkImage, label);
         }
 
