--- a/com/mojang/renderpearl/backend/vulkan/VulkanInstance.java
+++ b/com/mojang/renderpearl/backend/vulkan/VulkanInstance.java
@@ -2,7 +_,6 @@
 
 import com.mojang.logging.LogUtils;
 import com.mojang.renderpearl.api.device.BackendCreationException;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.nio.IntBuffer;
 import java.util.ArrayList;
 import java.util.HashSet;
@@ -26,7 +_,7 @@
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
-public class VulkanInstance implements UncheckedAutoCloseable {
+public class VulkanInstance implements AutoCloseable {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final String APPLICATION_NAME = "Minecraft Java Edition";
     private static final int APPLICATION_VERSION = SharedConstants.getCurrentVersion().dataVersion().version();
