--- a/com/mojang/renderpearl/backend/vulkan/VulkanTransientMemory.java
+++ b/com/mojang/renderpearl/backend/vulkan/VulkanTransientMemory.java
@@ -4,9 +_,11 @@
 import com.mojang.renderpearl.api.buffers.GpuBufferSlice;
 import com.mojang.renderpearl.api.buffers.TransientMemory;
 import com.mojang.renderpearl.backend.util.TransientBlockAllocator;
+import it.unimi.dsi.fastutil.Pair;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 import it.unimi.dsi.fastutil.ints.IntComparator;
 import it.unimi.dsi.fastutil.objects.ReferenceArrayList;
+import it.unimi.dsi.fastutil.objects.ReferenceReferenceImmutablePair;
 import java.nio.ByteBuffer;
 import java.nio.LongBuffer;
 import java.util.List;
@@ -20,7 +_,6 @@
 import org.lwjgl.system.MemoryUtil;
 import org.lwjgl.util.vma.Vma;
 import org.lwjgl.util.vma.VmaAllocationCreateInfo;
-import org.lwjgl.util.vma.VmaAllocationInfo;
 import org.lwjgl.vulkan.VK12;
 import org.lwjgl.vulkan.VkBufferCopy;
 import org.lwjgl.vulkan.VkBufferCreateInfo;
@@ -40,15 +_,12 @@
     private final VulkanDevice device;
     private final VulkanCommandEncoder encoder;
     private final boolean useDeviceMemoryForMappedGpuStaging;
-    private final TransientBlockAllocator<TransientBlockAllocator.Allocator.CpuBlock> cpuBlockAllocator = new TransientBlockAllocator<>(
-        524288L, 16L, TransientBlockAllocator.Allocator.CpuBlock.memalloc()
+    private final TransientBlockAllocator<Long> cpuBlockAllocator = new TransientBlockAllocator<>(
+        524288L, 16L, TransientBlockAllocator.Allocator.create(MemoryUtil::nmemAlloc, MemoryUtil::nmemFree)
     );
     private final TransientBlockAllocator<VulkanTransientMemory.VulkanAllocation> stagingBlockAllocator;
     private final TransientBlockAllocator<VulkanTransientMemory.VulkanAllocation> gpuBlockAllocator;
-    private final TransientBlockAllocator<VulkanTransientMemory.TransferPair> gpuMappedBlockAllocator;
-    private final int expectedCpuMemoryHeap;
-    private final int expectedGpuMemoryHeap;
-    private final int[] memoryTypeToHeapMap = new int[32];
+    private final TransientBlockAllocator<Pair<VulkanTransientMemory.VulkanAllocation, VulkanTransientMemory.VulkanAllocation>> gpuMappedBlockAllocator;
     private long submitIndex = 0L;
     private boolean anyCommandRecorded = false;
     private @Nullable VkCommandBuffer commandBuffer;
@@ -63,11 +_,6 @@
             VK12.vkGetPhysicalDeviceMemoryProperties(vkDevice.getPhysicalDevice(), memoryProperties);
             int heapCount = memoryProperties.memoryHeapCount();
             int typeCount = memoryProperties.memoryTypeCount();
-
-            for (int i = 0; i < typeCount; i++) {
-                this.memoryTypeToHeapMap[i] = memoryProperties.memoryTypes(i).heapIndex();
-            }
-
             int largestDeviceLocalHeapIndex = -1;
             long largestDeviceLocalHeapSize = -1L;
 
@@ -91,28 +_,6 @@
             }
 
             this.useDeviceMemoryForMappedGpuStaging = largestHeapIsHostVisibleAndCoherent;
-            int selectedCpuHeap = -1;
-            if (this.useDeviceMemoryForMappedGpuStaging) {
-                selectedCpuHeap = largestDeviceLocalHeapIndex;
-            } else {
-                for (int i = 0; i < typeCount; i++) {
-                    VkMemoryType typeProperties = memoryProperties.memoryTypes(i);
-                    if (VulkanUtils.hasAllBits(typeProperties.propertyFlags(), 6)) {
-                        VkMemoryHeap heapProperties = memoryProperties.memoryHeaps(typeProperties.heapIndex());
-                        if (!VulkanUtils.hasAnyBit(heapProperties.flags(), 1)) {
-                            selectedCpuHeap = typeProperties.heapIndex();
-                            break;
-                        }
-                    }
-                }
-            }
-
-            if (selectedCpuHeap == -1) {
-                throw new IllegalStateException("Could not select heap for CPU allocations");
-            }
-
-            this.expectedCpuMemoryHeap = selectedCpuHeap;
-            this.expectedGpuMemoryHeap = largestDeviceLocalHeapIndex;
         }
 
         this.stagingBlockAllocator = new TransientBlockAllocator<>(
@@ -166,17 +_,17 @@
         this.submitIndex++;
     }
 
-    private void recordGpuMappedCopy(VulkanTransientMemory.TransferPair block) {
-        if (block.cpu() != block.gpu()) {
-            assert block.cpu().size == block.gpu().size;
+    private void recordGpuMappedCopy(Pair<VulkanTransientMemory.VulkanAllocation, VulkanTransientMemory.VulkanAllocation> block) {
+        if (block.first() != block.second()) {
+            assert block.first().size == block.second().size;
 
             try (MemoryStack stack = MemoryStack.stackPush()) {
                 Buffer region = VkBufferCopy.calloc(1, stack);
                 region.srcOffset(0L);
                 region.dstOffset(0L);
-                region.size(block.cpu().size);
+                region.size(block.first().size);
                 assert this.commandBuffer != null;
-                VK12.vkCmdCopyBuffer(this.commandBuffer, block.cpu().vkBuffer, block.gpu().vkBuffer, region);
+                VK12.vkCmdCopyBuffer(this.commandBuffer, block.first().vkBuffer, block.second().vkBuffer, region);
                 this.anyCommandRecorded = true;
             }
         }
@@ -190,13 +_,10 @@
             bufferCreateInfo.sharingMode(0);
             bufferCreateInfo.pQueueFamilyIndices(null);
             VmaAllocationCreateInfo allocCreateInfo = VmaAllocationCreateInfo.calloc(stack);
-            int expectedHeap;
-            if (staging && !this.useDeviceMemoryForMappedGpuStaging) {
+            if (staging) {
                 allocCreateInfo.usage(9);
-                expectedHeap = this.expectedCpuMemoryHeap;
             } else {
                 allocCreateInfo.usage(8);
-                expectedHeap = this.expectedGpuMemoryHeap;
             }
 
             if (this.useDeviceMemoryForMappedGpuStaging || staging) {
@@ -206,8 +_,7 @@
 
             LongBuffer bufferPtr = stack.callocLong(1);
             PointerBuffer allocPtr = stack.callocPointer(1);
-            VmaAllocationInfo allocationInfo = VmaAllocationInfo.calloc(stack);
-            int result = Vma.vmaCreateBuffer(this.device.vma(), bufferCreateInfo, allocCreateInfo, bufferPtr, allocPtr, allocationInfo);
+            int result = Vma.vmaCreateBuffer(this.device.vma(), bufferCreateInfo, allocCreateInfo, bufferPtr, allocPtr, null);
             VulkanUtils.crashIfFailure(this.device, result, "Failed to allocate VkBuffer");
             PointerBuffer hostPtrPtr = stack.callocPointer(1);
             if (staging || this.useDeviceMemoryForMappedGpuStaging) {
@@ -215,8 +_,7 @@
             }
 
             this.device.instance().debug().setObjectName(this.device.vkDevice(), 9, bufferPtr.get(0), "Vulkan Transient Memory Buffer");
-            int allocatedHeap = this.memoryTypeToHeapMap[allocationInfo.memoryType()];
-            return new VulkanTransientMemory.VulkanAllocation(bufferPtr.get(0), allocPtr.get(0), hostPtrPtr.get(0), size, allocatedHeap != expectedHeap);
+            return new VulkanTransientMemory.VulkanAllocation(bufferPtr.get(0), allocPtr.get(0), hostPtrPtr.get(0), size);
         }
     }
 
@@ -228,30 +_,29 @@
         Vma.vmaDestroyBuffer(this.device.vma(), allocation.vkBuffer, allocation.vmaAllocation);
     }
 
-    private VulkanTransientMemory.TransferPair allocateGpuMappedVulkanBlock(long size) {
+    private Pair<VulkanTransientMemory.VulkanAllocation, VulkanTransientMemory.VulkanAllocation> allocateGpuMappedVulkanBlock(long size) {
         assert size >= 524288L;
         assert size >= this.gpuBlockAllocator.blockSize();
         if (this.useDeviceMemoryForMappedGpuStaging) {
             TransientBlockAllocator.Allocation<VulkanTransientMemory.VulkanAllocation> block = this.gpuBlockAllocator.allocate(size, 16L, size, 1L);
             assert block.offset() == 0L;
-            return new VulkanTransientMemory.TransferPair(block.block(), block.block());
+            return new ReferenceReferenceImmutablePair<>(block.block(), block.block());
         } else {
             assert size >= this.stagingBlockAllocator.blockSize();
             TransientBlockAllocator.Allocation<VulkanTransientMemory.VulkanAllocation> stagingBlock = this.stagingBlockAllocator.allocate(size, 16L, size, 1L);
             TransientBlockAllocator.Allocation<VulkanTransientMemory.VulkanAllocation> gpuBlock = this.gpuBlockAllocator.allocate(size, 16L, size, 1L);
-            return new VulkanTransientMemory.TransferPair(stagingBlock.block(), gpuBlock.block());
+            return new ReferenceReferenceImmutablePair<>(stagingBlock.block(), gpuBlock.block());
         }
     }
 
-    private void freeGpuMappedVulkanBlock(VulkanTransientMemory.TransferPair allocations) {
+    private void freeGpuMappedVulkanBlock(Pair<VulkanTransientMemory.VulkanAllocation, VulkanTransientMemory.VulkanAllocation> allocations) {
     }
 
     @Override
     public ByteBuffer allocateCpu(long size, long alignment, long minimumAllocation, long elementSize) {
         assert size <= 2147483647L;
-        TransientBlockAllocator.Allocation<TransientBlockAllocator.Allocator.CpuBlock> alloc = this.cpuBlockAllocator
-            .allocate(size, alignment, minimumAllocation, elementSize);
-        return MemoryUtil.memByteBuffer(alloc.block().address() + alloc.offset(), (int)alloc.size());
+        TransientBlockAllocator.Allocation<Long> alloc = this.cpuBlockAllocator.allocate(size, alignment, minimumAllocation, elementSize);
+        return MemoryUtil.memByteBuffer(alloc.block() + alloc.offset(), (int)alloc.size());
     }
 
     @Override
@@ -280,12 +_,12 @@
     @Override
     public GpuBufferSlice.MappedView allocateGpuMapped(long size, long alignment, @GpuBuffer.Usage int usage, long minimumAllocation, long elementSize) {
         assert size <= 2147483647L;
-        TransientBlockAllocator.Allocation<VulkanTransientMemory.TransferPair> alloc = this.gpuMappedBlockAllocator
+        TransientBlockAllocator.Allocation<Pair<VulkanTransientMemory.VulkanAllocation, VulkanTransientMemory.VulkanAllocation>> alloc = this.gpuMappedBlockAllocator
             .allocate(size, alignment, minimumAllocation, elementSize);
         VulkanTransientMemory.TransientGpuBuffer apiBuffer = new VulkanTransientMemory.TransientGpuBuffer(
-            alloc.block().gpu().vkBuffer, usage, (int)alloc.block().gpu().size, this.submitIndex
+            alloc.block().second().vkBuffer, usage, (int)alloc.block().first().size, this.submitIndex
         );
-        ByteBuffer cpuBuffer = MemoryUtil.memByteBuffer(alloc.block().cpu().hostPtr + alloc.offset(), (int)alloc.size());
+        ByteBuffer cpuBuffer = MemoryUtil.memByteBuffer(alloc.block().first().hostPtr + alloc.offset(), (int)alloc.size());
         return new GpuBufferSlice.MappedView(new GpuBufferSlice(apiBuffer, alloc.offset(), alloc.size()), cpuBuffer, () -> {});
     }
 
@@ -378,14 +_,6 @@
         return uploadedBuffers;
     }
 
-    private record TransferPair(VulkanTransientMemory.VulkanAllocation cpu, VulkanTransientMemory.VulkanAllocation gpu)
-        implements TransientBlockAllocator.Allocator.Block {
-        @Override
-        public boolean suboptimal() {
-            return this.cpu.suboptimal() || this.gpu.suboptimal();
-        }
-    }
-
     private class TransientGpuBuffer extends VulkanGpuBuffer {
         private boolean closed = false;
         private final long bufferSubmitIndex;
@@ -430,7 +_,6 @@
         }
     }
 
-    private record VulkanAllocation(long vkBuffer, long vmaAllocation, long hostPtr, long size, boolean suboptimal)
-        implements TransientBlockAllocator.Allocator.Block {
+    private record VulkanAllocation(long vkBuffer, long vmaAllocation, long hostPtr, long size) {
     }
 }
