--- a/com/mojang/renderpearl/backend/vulkan/VulkanConst.java
+++ b/com/mojang/renderpearl/backend/vulkan/VulkanConst.java
@@ -9,7 +_,6 @@
 import com.mojang.renderpearl.api.pipeline.CompareOp;
 import com.mojang.renderpearl.api.pipeline.PolygonMode;
 import com.mojang.renderpearl.api.pipeline.PrimitiveTopology;
-import com.mojang.renderpearl.api.pipeline.ShaderType;
 import com.mojang.renderpearl.api.textures.AddressMode;
 import com.mojang.renderpearl.api.textures.FilterMode;
 import com.mojang.renderpearl.api.textures.GpuTexture;
@@ -260,13 +_,6 @@
             case MAILBOX -> 1;
             case FIFO -> 2;
             case FIFO_RELAXED -> 3;
-        };
-    }
-
-    public static int toVk(ShaderType shaderType) {
-        return switch (shaderType) {
-            case VERTEX -> 1;
-            case FRAGMENT -> 16;
         };
     }
 }
