--- a/com/mojang/renderpearl/backend/vulkan/VulkanFeatureSets.java
+++ b/com/mojang/renderpearl/backend/vulkan/VulkanFeatureSets.java
@@ -30,6 +_,7 @@
         Set.of(
             new VulkanFeature(VK10_FEATURES_STRUCT, "multiDrawIndirect"),
             new VulkanFeature(VK10_FEATURES_STRUCT, "drawIndirectFirstInstance"),
+            new VulkanFeature(VK10_FEATURES_STRUCT, "fillModeNonSolid"),
             new VulkanFeature(VK10_FEATURES_STRUCT, "samplerAnisotropy"),
             new VulkanFeature(VK11_FEATURES_STRUCT, "shaderDrawParameters"),
             new VulkanFeature(VK12_FEATURES_STRUCT, "timelineSemaphore"),
@@ -51,10 +_,6 @@
     public static final FeatureSet NV_DIAGNOSTIC_CHECKPOINT_FEATURESET = new FeatureSet(
         "NV Diagnostic Checkpoint", Set.of("VK_NV_device_diagnostic_checkpoints"), Set.of(), device -> !AMD_BUFFER_MARKER_FEATURESET.isSupported(device)
     );
-    public static final FeatureSet WIREFRAME_FEATURESET = new FeatureSet(
-        "Wireframe", Set.of(), Set.of(new VulkanFeature(VK10_FEATURES_STRUCT, "fillModeNonSolid"))
-    );
-    public static final FeatureSet CALIBRATED_TIMESTAMP_FEATURESET = new FeatureSet("Calibrated Timestamps", Set.of("VK_EXT_calibrated_timestamps"), Set.of());
 
     public static Set<FeatureSet> requiredFeatureSets() {
         return ObjectOpenHashSet.of(REQUIRED_FEATURESET);
@@ -65,9 +_,7 @@
     }
 
     public static Set<FeatureSet> optionalFeatureSets() {
-        return ObjectOpenHashSet.of(
-            CALIBRATED_TIMESTAMP_FEATURESET, WIREFRAME_FEATURESET, NV_DIAGNOSTIC_CHECKPOINT_FEATURESET, AMD_BUFFER_MARKER_FEATURESET, MULTI_DRAW_FEATURESET
-        );
+        return ObjectOpenHashSet.of(NV_DIAGNOSTIC_CHECKPOINT_FEATURESET, AMD_BUFFER_MARKER_FEATURESET, MULTI_DRAW_FEATURESET);
     }
 
     public static void bootstrap() {
