--- a/com/mojang/renderpearl/backend/vulkan/VulkanQueue.java
+++ b/com/mojang/renderpearl/backend/vulkan/VulkanQueue.java
@@ -1,6 +_,5 @@
 package com.mojang.renderpearl.backend.vulkan;
 
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import it.unimi.dsi.fastutil.objects.ReferenceArrayList;
 import java.util.List;
 import net.neoforged.api.distmarker.Dist;
@@ -38,7 +_,7 @@
         VK12.vkQueueWaitIdle(this.vkQueue);
     }
 
-    public class Submission implements UncheckedAutoCloseable {
+    public class Submission implements AutoCloseable {
         private boolean closed = false;
         private final ReferenceArrayList<VulkanQueue.Submission.SubmitStage> stages = new ReferenceArrayList<>();
 
