--- a/com/mojang/renderpearl/backend/vulkan/VulkanPhysicalDevice.java
+++ b/com/mojang/renderpearl/backend/vulkan/VulkanPhysicalDevice.java
@@ -2,7 +_,6 @@
 
 import com.mojang.renderpearl.api.device.BackendCreationException;
 import com.mojang.renderpearl.api.device.DeviceType;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import it.unimi.dsi.fastutil.ints.Int2IntArrayMap;
 import it.unimi.dsi.fastutil.ints.Int2IntMap;
 import it.unimi.dsi.fastutil.ints.Int2IntMaps;
@@ -31,7 +_,7 @@
 import org.lwjgl.vulkan.VkExtensionProperties.Buffer;
 
 @OnlyIn(Dist.CLIENT)
-public class VulkanPhysicalDevice implements UncheckedAutoCloseable {
+public class VulkanPhysicalDevice implements AutoCloseable {
     private final VkPhysicalDevice vkPhysicalDevice;
     private final Buffer vkDeviceExtensions;
     private final VkPhysicalDeviceFeatures2 vkPhysicalDeviceFeatures;
