--- a/com/mojang/renderpearl/backend/vulkan/VulkanDevice.java
+++ b/com/mojang/renderpearl/backend/vulkan/VulkanDevice.java
@@ -1,48 +_,48 @@
 package com.mojang.renderpearl.backend.vulkan;
 
+import com.mojang.blaze3d.preprocessor.GlslPreprocessor;
 import com.mojang.logging.LogUtils;
 import com.mojang.renderpearl.api.GpuFormat;
 import com.mojang.renderpearl.api.buffers.GpuBuffer;
-import com.mojang.renderpearl.api.buffers.GpuBufferSlice;
 import com.mojang.renderpearl.api.commands.GpuQueryPool;
 import com.mojang.renderpearl.api.device.DeviceFeatures;
 import com.mojang.renderpearl.api.device.DeviceInfo;
 import com.mojang.renderpearl.api.device.DeviceLimits;
 import com.mojang.renderpearl.api.device.HintsAndWorkarounds;
+import com.mojang.renderpearl.api.pipeline.CompiledRenderPipeline;
+import com.mojang.renderpearl.api.pipeline.RenderPipeline;
+import com.mojang.renderpearl.api.pipeline.ShaderSource;
+import com.mojang.renderpearl.api.pipeline.ShaderType;
 import com.mojang.renderpearl.api.textures.AddressMode;
 import com.mojang.renderpearl.api.textures.FilterMode;
 import com.mojang.renderpearl.api.textures.GpuSampler;
 import com.mojang.renderpearl.api.textures.GpuTexture;
 import com.mojang.renderpearl.api.textures.GpuTextureView;
-import com.mojang.renderpearl.backend.api.BackendRenderPipeline;
 import com.mojang.renderpearl.backend.api.GpuDeviceBackend;
 import com.mojang.renderpearl.backend.api.GpuSurfaceBackend;
 import com.mojang.renderpearl.backend.vulkan.checkpoints.CheckpointExtension;
+import com.mojang.renderpearl.backend.vulkan.glsl.GlslCompiler;
+import com.mojang.renderpearl.backend.vulkan.glsl.IntermediaryShaderModule;
+import com.mojang.renderpearl.backend.vulkan.glsl.ShaderCompileException;
 import com.mojang.renderpearl.backend.vulkan.init.FeatureSet;
 import it.unimi.dsi.fastutil.ints.IntIntPair;
 import java.nio.ByteBuffer;
-import java.nio.LongBuffer;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
 import java.util.OptionalDouble;
 import java.util.Set;
-import java.util.function.BooleanSupplier;
 import java.util.function.Supplier;
-import net.minecraft.util.Util;
+import net.minecraft.client.renderer.ShaderDefines;
+import net.minecraft.resources.Identifier;
 import net.neoforged.api.distmarker.Dist;
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
-import org.lwjgl.system.MemoryStack;
 import org.lwjgl.util.vma.Vma;
-import org.lwjgl.vulkan.EXTCalibratedTimestamps;
 import org.lwjgl.vulkan.VK12;
-import org.lwjgl.vulkan.VkBufferCopy;
-import org.lwjgl.vulkan.VkCalibratedTimestampInfoEXT;
 import org.lwjgl.vulkan.VkDevice;
 import org.lwjgl.vulkan.VkPhysicalDeviceLimits;
 import org.lwjgl.vulkan.VkPhysicalDeviceVulkan11Properties;
-import org.lwjgl.vulkan.VkBufferCopy.Buffer;
 import org.slf4j.Logger;
 
 @OnlyIn(Dist.CLIENT)
@@ -51,12 +_,12 @@
     private final VulkanInstance instance;
     private final VkDevice vkDevice;
     private final long vma;
+    private final GlslCompiler glslCompiler = new GlslCompiler();
     private final DeviceInfo deviceInfo;
     private final VulkanQueue graphicsQueue;
     private final VulkanQueue computeQueue;
     private final VulkanQueue transferQueue;
     private final boolean isIntegratedIntelMoltenVK;
-    private final FeatureSet enabledFeatures;
     private final VulkanCommandEncoder commandEncoder;
     private final CheckpointExtension checkpointExtension;
 
@@ -71,7 +_,6 @@
         this.instance = instance;
         this.vkDevice = vkDevice;
         this.vma = vma;
-        this.enabledFeatures = enabledFeatureSet;
         this.checkpointExtension = checkpointExtension;
         Set<String> extensionNames = new HashSet<>();
 
@@ -85,9 +_,6 @@
 
         VkPhysicalDeviceLimits limits = physicalDevice.vkPhysicalDeviceProperties().limits();
         VkPhysicalDeviceVulkan11Properties vk11Properties = physicalDevice.vkPhysicalDeviceVulkan11Properties();
-        int indirectDrawCount = Integer.compareUnsigned(limits.maxDrawIndirectCount(), Integer.MAX_VALUE) > 0
-            ? Integer.MAX_VALUE
-            : limits.maxDrawIndirectCount();
         this.deviceInfo = new DeviceInfo(
             physicalDevice.deviceName(),
             physicalDevice.vendorName(),
@@ -103,21 +_,11 @@
                 physicalDevice.vkPhysicalDeviceMultiDrawPropertiesEXT().maxMultiDrawCount() < 0
                     ? Integer.MAX_VALUE
                     : physicalDevice.vkPhysicalDeviceMultiDrawPropertiesEXT().maxMultiDrawCount(),
-                limits.maxColorAttachments(),
-                indirectDrawCount
-            ),
-            new DeviceFeatures(
-                enabledFeatureSet.contains(VulkanFeatureSets.WIREFRAME_FEATURESET),
-                true,
-                enabledFeatureSet.contains(VulkanFeatureSets.MULTI_DRAW_FEATURESET),
-                false,
-                true,
-                true,
-                true,
-                true
-            ),
+                limits.maxColorAttachments()
+            ),
+            new DeviceFeatures(true, enabledFeatureSet.contains(VulkanFeatureSets.MULTI_DRAW_FEATURESET), false, true, true, true, true),
             Collections.unmodifiableSet(extensionNames),
-            new HintsAndWorkarounds(false, false, Util.isAppleSiliconMac(physicalDevice.deviceName()), false),
+            new HintsAndWorkarounds(false, false),
             physicalDevice.deviceType()
         );
         IntIntPair graphicsQueueFamily = physicalDevice.graphicsQueueFamilyAndIndex();
@@ -151,6 +_,7 @@
         Vma.vmaDestroyAllocator(this.vma);
         VK12.vkDestroyDevice(this.vkDevice, null);
         this.instance.close();
+        this.glslCompiler.close();
     }
 
     @Override
@@ -183,7 +_,7 @@
     }
 
     @Override
-    public GpuSurfaceBackend createSurface(long windowHandle, BooleanSupplier isIconified) {
+    public GpuSurfaceBackend createSurface(long windowHandle) {
         return new VulkanGpuSurface(this, windowHandle);
     }
 
@@ -200,12 +_,24 @@
 
     @Override
     public GpuTexture createTexture(
+        @Nullable Supplier<String> label, @GpuTexture.Usage int usage, GpuFormat format, int width, int height, int depthOrLayers, int mipLevels
+    ) {
+        return new VulkanGpuTexture(this, usage, this.isDebuggingEnabled() && label != null ? label.get() : "", format, width, height, depthOrLayers, mipLevels);
+    }
+
+    @Override
+    public GpuTexture createTexture(
         @Nullable String label, @GpuTexture.Usage int usage, GpuFormat format, int width, int height, int depthOrLayers, int mipLevels
     ) {
         return new VulkanGpuTexture(this, usage, this.isDebuggingEnabled() && label != null ? label : "", format, width, height, depthOrLayers, mipLevels);
     }
 
     @Override
+    public GpuTextureView createTextureView(GpuTexture texture) {
+        return this.createTextureView(texture, 0, texture.getMipLevels());
+    }
+
+    @Override
     public GpuTextureView createTextureView(GpuTexture texture, int baseMipLevel, int mipLevels) {
         return new VulkanGpuTextureView(this, (VulkanGpuTexture)texture, baseMipLevel, mipLevels);
     }
@@ -216,16 +_,8 @@
 
     @Override
     public GpuBuffer createBuffer(@Nullable Supplier<String> label, @GpuBuffer.Usage int usage, ByteBuffer data) {
-        VulkanGpuBuffer buffer = this.createBuffer(label, usage | 8, data.remaining());
-        GpuBufferSlice stagingBuffer = this.commandEncoder.transientMemory().uploadStaging(data, 1L, 16);
-
-        try (MemoryStack stack = MemoryStack.stackPush()) {
-            Buffer regions = VkBufferCopy.calloc(1, stack).srcOffset(stagingBuffer.offset()).dstOffset(0L).size(data.remaining());
-            VK12.vkCmdCopyBuffer(
-                this.commandEncoder.objectInitCommandBuffer(), ((VulkanGpuBuffer)stagingBuffer.buffer()).vkBuffer(), buffer.vkBuffer(), regions
-            );
-        }
-
+        GpuBuffer buffer = this.createBuffer(label, usage | 8, data.remaining());
+        this.createCommandEncoder().writeToBuffer(buffer.slice(), data);
         return buffer;
     }
 
@@ -239,10 +_,49 @@
         return this.instance.debug().enabled();
     }
 
+    protected @Nullable IntermediaryShaderModule compileShader(Identifier id, ShaderType type, ShaderDefines defines, ShaderSource shaderSource) {
+        String source = shaderSource.get(id, type);
+        if (source == null) {
+            LOGGER.error("Couldn't find source for {} shader ({})", type, id);
+            return null;
+        }
+
+        String sourceWithDefines = GlslPreprocessor.injectDefines(source, defines);
+
+        try {
+            return this.glslCompiler.createIntermediary(id.toDebugFileName(), sourceWithDefines, type);
+        } catch (ShaderCompileException e) {
+            LOGGER.error("Couldn't compile {} shader {}", type, id, e);
+            return null;
+        }
+    }
+
     @Override
-    public BackendRenderPipeline.Pending compilePipeline(BackendRenderPipeline.CreateInfo pipelineCreateInfo) {
-        VulkanRenderPipeline pipeline = VulkanRenderPipeline.compile(this, pipelineCreateInfo);
-        return () -> pipeline;
+    public @Nullable CompiledRenderPipeline compilePipeline(RenderPipeline pipeline, ShaderSource shaderSource) {
+        IntermediaryShaderModule vertexShader = this.compileShader(pipeline.getVertexShader(), ShaderType.VERTEX, pipeline.getShaderDefines(), shaderSource);
+        if (vertexShader == null) {
+            LOGGER.error("Couldn't compile pipeline {}: vertex shader {} was invalid", pipeline.getLocation(), pipeline.getVertexShader());
+            return null;
+        }
+
+        IntermediaryShaderModule fragmentShader = this.compileShader(
+            pipeline.getFragmentShader(), ShaderType.FRAGMENT, pipeline.getShaderDefines(), shaderSource
+        );
+        if (fragmentShader == null) {
+            LOGGER.error("Couldn't compile pipeline {}: fragment shader {} was invalid", pipeline.getLocation(), pipeline.getFragmentShader());
+            vertexShader.close();
+            return null;
+        }
+
+        try {
+            GlslCompiler.CompiledModules modules = this.glslCompiler.compile(this, pipeline, vertexShader, fragmentShader);
+            return VulkanRenderPipeline.compile(this, modules.layout(), pipeline, modules.vertex(), modules.fragment());
+        } catch (ShaderCompileException e) {
+            LOGGER.error("Couldn't compile pipeline {}", pipeline.getLocation(), e);
+            vertexShader.close();
+            fragmentShader.close();
+            return null;
+        }
     }
 
     @Override
@@ -251,27 +_,8 @@
     }
 
     @Override
-    public long getTimestampCalibrationOffset() {
-        double timestampPeriod = this.deviceInfo.timestampPeriod();
-        if (!this.enabledFeatures.contains(VulkanFeatureSets.CALIBRATED_TIMESTAMP_FEATURESET)) {
-            long deviceTime = this.commandEncoder.getTimestampNow();
-            long hostTime = System.nanoTime();
-            long deviceTimeInNanos = timestampPeriod == 1.0 ? deviceTime : (long)(deviceTime * timestampPeriod);
-            return hostTime - deviceTimeInNanos;
-        }
-
-        try (MemoryStack stack = MemoryStack.stackPush()) {
-            org.lwjgl.vulkan.VkCalibratedTimestampInfoEXT.Buffer infos = VkCalibratedTimestampInfoEXT.calloc(2, stack);
-            infos.get(0).sType$Default().timeDomain(0);
-            infos.get(1).sType$Default().timeDomain(1);
-            LongBuffer timestampValues = stack.callocLong(infos.capacity());
-            LongBuffer deviation = stack.callocLong(1);
-            EXTCalibratedTimestamps.vkGetCalibratedTimestampsEXT(this.vkDevice, infos, timestampValues, deviation);
-            long deviceTime = timestampValues.get(0);
-            long hostTime = timestampValues.get(1);
-            long deviceInNanos = timestampPeriod == 1.0 ? deviceTime : (long)(deviceTime * timestampPeriod);
-            return hostTime - deviceInNanos;
-        }
+    public long getTimestampNow() {
+        return this.commandEncoder.getTimestampNow();
     }
 
     public CheckpointExtension checkpointExtension() {
