--- a/net/minecraft/network/chat/contents/EntityDataSource.java
+++ b/net/minecraft/network/chat/contents/EntityDataSource.java
@@ -5,7 +_,6 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.List;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
