--- a/net/minecraft/network/chat/contents/EntityDataSource.java
+++ b/net/minecraft/network/chat/contents/EntityDataSource.java
@@ -29,7 +_,7 @@
     @Nullable
     private static EntitySelector compileSelector(String p_237336_) {
         try {
-            EntitySelectorParser entityselectorparser = new EntitySelectorParser(new StringReader(p_237336_), true);
+            EntitySelectorParser entityselectorparser = new EntitySelectorParser(new StringReader(p_237336_));
             return entityselectorparser.parse();
         } catch (CommandSyntaxException commandsyntaxexception) {
             return null;
