--- a/net/minecraft/network/chat/contents/NbtContents.java
+++ b/net/minecraft/network/chat/contents/NbtContents.java
@@ -7,14 +_,12 @@
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.Optional;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.arguments.NbtPathArgument;
-import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.Tag;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentContents;
@@ -122,7 +_,7 @@
             }).map(Tag::getAsString);
             if (this.interpreting) {
                 Component component = DataFixUtils.orElse(
-                    ComponentUtils.updateForEntity(p_237401_, this.separator, p_237402_, p_237403_), ComponentUtils.DEFAULT_NO_STYLE_SEPARATOR
+                    ComponentUtils.updateSeparatorForEntity(p_237401_, this.separator, p_237402_, p_237403_), ComponentUtils.DEFAULT_NO_STYLE_SEPARATOR // Paper - validate separator
                 );
                 return stream.flatMap(p_329869_ -> {
                     try {
@@ -134,7 +_,7 @@
                     }
                 }).reduce((p_237420_, p_237421_) -> p_237420_.append(component).append(p_237421_)).orElseGet(Component::empty);
             } else {
-                return ComponentUtils.updateForEntity(p_237401_, this.separator, p_237402_, p_237403_)
+                return ComponentUtils.updateSeparatorForEntity(p_237401_, this.separator, p_237402_, p_237403_) // Paper - validate separator
                     .map(
                         p_237415_ -> stream.map(Component::literal)
                                 .reduce((p_237424_, p_237425_) -> p_237424_.append(p_237415_).append(p_237425_))
