--- a/net/minecraft/network/chat/contents/ScoreContents.java
+++ b/net/minecraft/network/chat/contents/ScoreContents.java
@@ -40,7 +_,7 @@
     @Nullable
     private static EntitySelector parseSelector(String p_237448_) {
         try {
-            return new EntitySelectorParser(new StringReader(p_237448_), true).parse();
+            return new EntitySelectorParser(new StringReader(p_237448_)).parse();
         } catch (CommandSyntaxException commandsyntaxexception) {
             return null;
         }
