--- a/net/minecraft/network/chat/contents/SelectorContents.java
+++ b/net/minecraft/network/chat/contents/SelectorContents.java
@@ -48,7 +_,7 @@
         EntitySelector entityselector = null;
 
         try {
-            EntitySelectorParser entityselectorparser = new EntitySelectorParser(new StringReader(p_237472_), true);
+            EntitySelectorParser entityselectorparser = new EntitySelectorParser(new StringReader(p_237472_));
             entityselector = entityselectorparser.parse();
         } catch (CommandSyntaxException commandsyntaxexception) {
             LOGGER.warn("Invalid selector component: {}: {}", p_237472_, commandsyntaxexception.getMessage());
