--- a/net/minecraft/world/level/levelgen/feature/treedecorators/AttachedToLogsDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/AttachedToLogsDecorator.java
@@ -6,7 +_,6 @@
 import java.util.List;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
-import net.minecraft.core.Holder;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.Util;
@@ -22,10 +_,10 @@
             .apply(i, AttachedToLogsDecorator::new)
     );
     private final float probability;
-    private final Holder<BlockStateProvider> blockProvider;
+    private final BlockStateProvider blockProvider;
     private final List<Direction> directions;
 
-    public AttachedToLogsDecorator(float probability, Holder<BlockStateProvider> blockProvider, List<Direction> directions) {
+    public AttachedToLogsDecorator(float probability, BlockStateProvider blockProvider, List<Direction> directions) {
         this.probability = probability;
         this.blockProvider = blockProvider;
         this.directions = directions;
@@ -39,7 +_,7 @@
             Direction direction = Util.getRandom(this.directions, random);
             BlockPos placementPos = logsPos.relative(direction);
             if (random.nextFloat() <= this.probability && context.isAir(placementPos)) {
-                context.setBlock(placementPos, this.blockProvider.value().getState(context.level(), random, placementPos));
+                context.setBlock(placementPos, this.blockProvider.getState(context.level(), random, placementPos));
             }
         }
     }
