--- a/net/minecraft/world/level/levelgen/feature/treedecorators/AttachedToLeavesDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/AttachedToLeavesDecorator.java
@@ -8,7 +_,6 @@
 import java.util.Set;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
-import net.minecraft.core.Holder;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.Util;
@@ -29,17 +_,12 @@
     protected final float probability;
     protected final int exclusionRadiusXZ;
     protected final int exclusionRadiusY;
-    protected final Holder<BlockStateProvider> blockProvider;
+    protected final BlockStateProvider blockProvider;
     protected final int requiredEmptyBlocks;
     protected final List<Direction> directions;
 
     public AttachedToLeavesDecorator(
-        float probability,
-        int exclusionRadiusXZ,
-        int exclusionRadiusY,
-        Holder<BlockStateProvider> blockProvider,
-        int requiredEmptyBlocks,
-        List<Direction> directions
+        float probability, int exclusionRadiusXZ, int exclusionRadiusY, BlockStateProvider blockProvider, int requiredEmptyBlocks, List<Direction> directions
     ) {
         this.probability = probability;
         this.exclusionRadiusXZ = exclusionRadiusXZ;
@@ -67,7 +_,7 @@
                     propaguleBlacklist.add(inPos.immutable());
                 }
 
-                context.setBlock(placementPos, this.blockProvider.value().getState(context.level(), random, placementPos));
+                context.setBlock(placementPos, this.blockProvider.getState(context.level(), random, placementPos));
             }
         }
     }
