--- a/net/minecraft/world/level/levelgen/feature/treedecorators/PaleMossDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/PaleMossDecorator.java
@@ -51,7 +_,7 @@
                 level.registryAccess()
                     .lookup(Registries.CONFIGURED_FEATURE)
                     .flatMap(registry -> registry.get(VegetationFeatures.PALE_MOSS_PATCH))
-                    .ifPresent(mossPatch -> mossPatch.value().place(level, level.getLevel().getChunkSource().getGenerator(), random, origin.above()));
+                    .ifPresent(mossPatch -> mossPatch.value().place(level, ((net.minecraft.server.level.ServerChunkCache) level.getChunkSource()).getGenerator(), random, origin.above())); // CraftBukkit - don't use level from DummyGeneratorAccess
             }
 
             context.logs().forEach(pos -> {
