--- a/net/minecraft/world/level/levelgen/feature/treedecorators/PlaceOnGroundDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/PlaceOnGroundDecorator.java
@@ -4,7 +_,6 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import java.util.List;
 import net.minecraft.core.BlockPos;
-import net.minecraft.core.Holder;
 import net.minecraft.util.ExtraCodecs;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.block.Blocks;
@@ -27,9 +_,9 @@
     private final int tries;
     private final int radius;
     private final int height;
-    private final Holder<BlockStateProvider> blockStateProvider;
+    private final BlockStateProvider blockStateProvider;
 
-    public PlaceOnGroundDecorator(int tries, int radius, int height, Holder<BlockStateProvider> blockStateProvider) {
+    public PlaceOnGroundDecorator(int tries, int radius, int height, BlockStateProvider blockStateProvider) {
         this.tries = tries;
         this.radius = radius;
         this.height = height;
@@ -81,7 +_,7 @@
         if (context.level().isStateAtPosition(abovePos, state -> state.isAir() || state.is(Blocks.VINE))
             && context.checkBlock(pos, BlockBehaviour.BlockStateBase::isSolidRender)
             && context.level().getHeightmapPos(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES, pos).getY() <= abovePos.getY()) {
-            context.setBlock(abovePos, this.blockStateProvider.value().getState(context.level(), context.random(), abovePos));
+            context.setBlock(abovePos, this.blockStateProvider.getState(context.level(), context.random(), abovePos));
         }
     }
 }
