--- a/net/minecraft/ChatFormatting.java
+++ b/net/minecraft/ChatFormatting.java
@@ -112,6 +_,18 @@
         return p_126658_ == null ? null : FORMATTING_BY_NAME.get(cleanName(p_126658_));
     }
 
+    // Paper start - add method to get by hex value
+    @Nullable public static ChatFormatting getByHexValue(int i) {
+        for (ChatFormatting value : values()) {
+            if (value.getColor() != null && value.getColor() == i) {
+                return value;
+            }
+        }
+
+        return null;
+    }
+    // Paper end - add method to get by hex value
+
     @Nullable
     public static ChatFormatting getById(int p_126648_) {
         if (p_126648_ < 0) {
