--- a/net/minecraft/CrashReport.java
+++ b/net/minecraft/CrashReport.java
@@ -15,8 +_,8 @@
 import java.util.concurrent.CompletionException;
 import javax.annotation.Nullable;
 import net.minecraft.util.MemoryReserve;
-import org.apache.commons.io.IOUtils;
 import org.apache.commons.lang3.ArrayUtils;
+import org.bukkit.craftbukkit.CraftCrashReport;
 import org.slf4j.Logger;
 
 public class CrashReport {
@@ -30,10 +_,12 @@
     private boolean trackingStackTrace = true;
     private StackTraceElement[] uncategorizedStackTrace = new StackTraceElement[0];
     private final SystemReport systemReport = new SystemReport();
+    private List<String> extraInfo = List.of("", "DO NOT REPORT THIS TO PAPER! REPORT TO YOUER INSTEAD!", ""); // Purpur - Rebrand
 
     public CrashReport(String p_127509_, Throwable p_127510_) {
         this.title = p_127509_;
         this.exception = p_127510_;
+        this.systemReport.setDetail("CraftBukkit Information", new CraftCrashReport()); // CraftBukkit
     }
 
     public String getTitle() {
@@ -58,14 +_,8 @@
         if (this.uncategorizedStackTrace != null && this.uncategorizedStackTrace.length > 0) {
             p_127520_.append("-- Head --\n");
             p_127520_.append("Thread: ").append(Thread.currentThread().getName()).append("\n");
-            p_127520_.append("Stacktrace:\n");
-
-            for (StackTraceElement stacktraceelement : this.uncategorizedStackTrace) {
-                p_127520_.append("\t").append("at ").append(stacktraceelement);
-                p_127520_.append("\n");
-            }
-
-            p_127520_.append("\n");
+            p_127520_.append("Stacktrace:");
+            p_127520_.append(net.neoforged.neoforge.logging.CrashReportExtender.generateEnhancedStackTrace(this.uncategorizedStackTrace));
         }
 
         for (CrashReportCategory crashreportcategory : this.details) {
@@ -73,6 +_,7 @@
             p_127520_.append("\n\n");
         }
 
+        net.neoforged.neoforge.logging.CrashReportExtender.extendSystemReport(systemReport);
         this.systemReport.appendToCrashReportString(p_127520_);
     }
 
@@ -92,18 +_,7 @@
             throwable.setStackTrace(this.exception.getStackTrace());
         }
 
-        String s;
-        try {
-            stringwriter = new StringWriter();
-            printwriter = new PrintWriter(stringwriter);
-            throwable.printStackTrace(printwriter);
-            s = stringwriter.toString();
-        } finally {
-            IOUtils.closeQuietly((Writer)stringwriter);
-            IOUtils.closeQuietly((Writer)printwriter);
-        }
-
-        return s;
+        return net.neoforged.neoforge.logging.CrashReportExtender.generateEnhancedStackTrace(throwable);
     }
 
     public String getFriendlyReport(ReportType p_350860_, List<String> p_350563_) {
@@ -128,7 +_,7 @@
     }
 
     public String getFriendlyReport(ReportType p_351019_) {
-        return this.getFriendlyReport(p_351019_, List.of());
+        return this.getFriendlyReport(p_351019_, extraInfo); // Purpur - Rebrand
     }
 
     @Nullable
@@ -159,7 +_,7 @@
     }
 
     public boolean saveToFile(Path p_350691_, ReportType p_350920_) {
-        return this.saveToFile(p_350691_, p_350920_, List.of());
+        return this.saveToFile(p_350691_, p_350920_, extraInfo); // Purpur - Rebrand
     }
 
     public SystemReport getSystemReport() {
@@ -218,7 +_,7 @@
     }
 
     public static void preload() {
-        MemoryReserve.allocate();
+        // MemoryReserve.allocate(); // Paper - Disable memory reserve allocating
         new CrashReport("Don't panic!", new Throwable()).getFriendlyReport(ReportType.CRASH);
     }
 }
