--- a/net/minecraft/DetectedVersion.java
+++ b/net/minecraft/DetectedVersion.java
@@ -25,7 +_,7 @@
 
     public static WorldVersion createBuiltIn(String id, String name, boolean stable) {
         return new WorldVersion.Simple(
-            id, name, new DataVersion(5017, "main"), SharedConstants.getProtocolVersion(), PackFormat.of(97, 1), PackFormat.of(119, 0), new Date(), stable
+            id, name, new DataVersion(5003, "main"), SharedConstants.getProtocolVersion(), PackFormat.of(92, 0), PackFormat.of(111, 0), new Date(), stable
         );
     }
 
