--- a/net/minecraft/SharedConstants.java
+++ b/net/minecraft/SharedConstants.java
@@ -13,22 +_,22 @@
     @Deprecated
     public static final boolean SNAPSHOT = true;
     @Deprecated
-    public static final int WORLD_VERSION = 5017;
+    public static final int WORLD_VERSION = 5003;
     @Deprecated
     public static final String SERIES = "main";
     @Deprecated
     public static final int RELEASE_NETWORK_PROTOCOL_VERSION = 777;
     @Deprecated
-    public static final int SNAPSHOT_NETWORK_PROTOCOL_VERSION = 333;
+    public static final int SNAPSHOT_NETWORK_PROTOCOL_VERSION = 326;
     public static final int SNBT_NAG_VERSION = 4997;
     private static final int SNAPSHOT_PROTOCOL_BIT = 30;
     public static final boolean CRASH_EAGERLY = true;
     @Deprecated
-    public static final int RESOURCE_PACK_FORMAT_MAJOR = 97;
-    @Deprecated
-    public static final int RESOURCE_PACK_FORMAT_MINOR = 1;
-    @Deprecated
-    public static final int DATA_PACK_FORMAT_MAJOR = 119;
+    public static final int RESOURCE_PACK_FORMAT_MAJOR = 92;
+    @Deprecated
+    public static final int RESOURCE_PACK_FORMAT_MINOR = 0;
+    @Deprecated
+    public static final int DATA_PACK_FORMAT_MAJOR = 111;
     @Deprecated
     public static final int DATA_PACK_FORMAT_MINOR = 0;
     public static final String RPC_MANAGEMENT_SERVER_API_VERSION = "3.1.0";
@@ -120,13 +_,12 @@
     public static final boolean DEBUG_CALCULATE_SOLID = debugFlag("CALCULATE_SOLID");
     public static final boolean DEBUG_FORCE_TELEMETRY = debugFlag("FORCE_TELEMETRY");
     public static final boolean DEBUG_DONT_SEND_TELEMETRY_TO_BACKEND = debugFlag("DONT_SEND_TELEMETRY_TO_BACKEND");
-    public static final boolean DEBUG_ENABLE_FARLANDS = debugFlag("ENABLE_FARLANDS");
     public static final long MAXIMUM_TICK_TIME_NANOS = Duration.ofMillis(300L).toNanos();
     public static final float MAXIMUM_BLOCK_EXPLOSION_RESISTANCE = 3600000.0F;
     public static final boolean USE_DEVONLY = false;
     public static boolean CHECK_DATA_FIXER_SCHEMA = true;
-    public static boolean IS_RUNNING_IN_IDE;
-    public static boolean IS_RENDERDOC_ATTACHED = false;
+    public static boolean IS_RUNNING_IN_IDE = !net.neoforged.fml.loading.FMLEnvironment.isProduction();
+    public static final boolean IS_RUNNING_WITH_JDWP = java.lang.management.ManagementFactory.getRuntimeMXBean().getInputArguments().stream().anyMatch(str -> str.startsWith("-agentlib:jdwp"));
     public static final int WORLD_RESOLUTION = 16;
     public static final int MAX_CHAT_LENGTH = 256;
     public static final int MAX_USER_INPUT_COMMAND_LENGTH = 32500;
@@ -205,7 +_,7 @@
     }
 
     public static int getProtocolVersion() {
-        return 1073742157;
+        return 1073742150;
     }
 
     public static boolean debugVoidTerrain(ChunkPos pos) {
@@ -219,6 +_,7 @@
     }
 
     static {
+        if (System.getProperty("io.netty.leakDetection.level") == null) // Forge: allow level to be manually specified
         ResourceLeakDetector.setLevel(NETTY_LEAK_DETECTION);
         CommandSyntaxException.ENABLE_COMMAND_STACK_TRACES = COMMAND_STACK_TRACES;
         CommandSyntaxException.BUILT_IN_EXCEPTIONS = new BrigadierExceptions();
