--- a/net/minecraft/SharedConstants.java
+++ b/net/minecraft/SharedConstants.java
@@ -13,22 +_,22 @@
     @Deprecated
     public static final boolean SNAPSHOT = true;
     @Deprecated
-    public static final int WORLD_VERSION = 5003;
+    public static final int WORLD_VERSION = 4999;
     @Deprecated
     public static final String SERIES = "main";
     @Deprecated
     public static final int RELEASE_NETWORK_PROTOCOL_VERSION = 777;
     @Deprecated
-    public static final int SNAPSHOT_NETWORK_PROTOCOL_VERSION = 326;
+    public static final int SNAPSHOT_NETWORK_PROTOCOL_VERSION = 324;
     public static final int SNBT_NAG_VERSION = 4997;
     private static final int SNAPSHOT_PROTOCOL_BIT = 30;
     public static final boolean CRASH_EAGERLY = true;
     @Deprecated
-    public static final int RESOURCE_PACK_FORMAT_MAJOR = 92;
+    public static final int RESOURCE_PACK_FORMAT_MAJOR = 90;
     @Deprecated
     public static final int RESOURCE_PACK_FORMAT_MINOR = 0;
     @Deprecated
-    public static final int DATA_PACK_FORMAT_MAJOR = 111;
+    public static final int DATA_PACK_FORMAT_MAJOR = 109;
     @Deprecated
     public static final int DATA_PACK_FORMAT_MINOR = 0;
     public static final String RPC_MANAGEMENT_SERVER_API_VERSION = "3.1.0";
@@ -91,6 +_,7 @@
     public static final boolean DEBUG_VERBOSE_COMMAND_ERRORS = debugFlag("VERBOSE_COMMAND_ERRORS");
     public static final boolean DEBUG_DEV_COMMANDS = debugFlag("DEV_COMMANDS");
     public static final boolean DEBUG_ACTIVE_TEXT_AREAS = debugFlag("ACTIVE_TEXT_AREAS");
+    public static final boolean DEBUG_PREFER_WAYLAND = debugFlag("PREFER_WAYLAND");
     public static final boolean DEBUG_SIMULATE_LIBRARY_LOAD_FAILURE = debugFlag("SIMULATE_LIBRARY_LOAD_FAILURE");
     public static final boolean DEBUG_IGNORE_LOCAL_MOB_CAP = debugFlag("IGNORE_LOCAL_MOB_CAP");
     public static final boolean DEBUG_DISABLE_LIQUID_SPREADING = debugFlag("DISABLE_LIQUID_SPREADING");
@@ -124,7 +_,8 @@
     public static final float MAXIMUM_BLOCK_EXPLOSION_RESISTANCE = 3600000.0F;
     public static final boolean USE_DEVONLY = false;
     public static boolean CHECK_DATA_FIXER_SCHEMA = true;
-    public static boolean IS_RUNNING_IN_IDE;
+    public static boolean IS_RUNNING_IN_IDE = !net.neoforged.fml.loading.FMLEnvironment.isProduction();
+    public static final boolean IS_RUNNING_WITH_JDWP = java.lang.management.ManagementFactory.getRuntimeMXBean().getInputArguments().stream().anyMatch(str -> str.startsWith("-agentlib:jdwp"));
     public static final int WORLD_RESOLUTION = 16;
     public static final int MAX_CHAT_LENGTH = 256;
     public static final int MAX_USER_INPUT_COMMAND_LENGTH = 32500;
@@ -203,7 +_,7 @@
     }
 
     public static int getProtocolVersion() {
-        return 1073742150;
+        return 1073742148;
     }
 
     public static boolean debugVoidTerrain(ChunkPos pos) {
@@ -217,6 +_,7 @@
     }
 
     static {
+        if (System.getProperty("io.netty.leakDetection.level") == null) // Forge: allow level to be manually specified
         ResourceLeakDetector.setLevel(NETTY_LEAK_DETECTION);
         CommandSyntaxException.ENABLE_COMMAND_STACK_TRACES = COMMAND_STACK_TRACES;
         CommandSyntaxException.BUILT_IN_EXCEPTIONS = new BrigadierExceptions();
