--- a/net/minecraft/SystemReport.java
+++ b/net/minecraft/SystemReport.java
@@ -12,6 +_,7 @@
 import java.util.Map.Entry;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
+import org.bukkit.Bukkit;
 import org.slf4j.Logger;
 import oshi.SystemInfo;
 import oshi.hardware.CentralProcessor;
@@ -40,6 +_,7 @@
     private final Map<String, String> entries = Maps.newLinkedHashMap();
 
     public SystemReport() {
+        this.setDetail("Youer Version", io.papermc.paper.ServerBuildInfo.buildInfo().asString(io.papermc.paper.ServerBuildInfo.StringRepresentation.VERSION_SIMPLE));
         this.setDetail("Minecraft Version", SharedConstants.getCurrentVersion().getName());
         this.setDetail("Minecraft Version ID", SharedConstants.getCurrentVersion().getId());
         this.setDetail("Operating System", OPERATING_SYSTEM);
