--- a/net/minecraft/world/entity/animal/fish/WaterAnimal.java
+++ b/net/minecraft/world/entity/animal/fish/WaterAnimal.java
@@ -41,7 +_,7 @@
             this.setAirSupply(preTickAirSupply - 1);
             if (this.shouldTakeDrowningDamage()) {
                 this.setAirSupply(0);
-                this.hurtServer(level, this.damageSources().drown(), 2.0F);
+                this.hurtServer(level, this.damageSources().drown().knownCause(org.bukkit.event.entity.EntityDamageEvent.DamageCause.DRYOUT), 2.0F); // Paper
             }
         } else {
             this.setAirSupply(300);
@@ -72,8 +_,11 @@
     ) {
         int seaLevel = level.getSeaLevel();
         int minSpawnLevel = seaLevel - 13;
-        return pos.getY() >= minSpawnLevel
-            && pos.getY() <= seaLevel
+        // Paper start - Make water animal spawn height configurable
+        seaLevel = level.getMinecraftWorld().paperConfig().entities.spawning.wateranimalSpawnHeight.maximum.or(seaLevel);
+        minSpawnLevel = level.getMinecraftWorld().paperConfig().entities.spawning.wateranimalSpawnHeight.minimum.or(minSpawnLevel);
+        // Paper end - Make water animal spawn height configurable
+        return ((spawnReason == EntitySpawnReason.SPAWNER && level.getMinecraftWorld().purpurConfig.spawnerFixMC238526) || (pos.getY() >= minSpawnLevel && pos.getY() <= seaLevel)) // Purpur - MC-238526 - Fix spawner not spawning water animals correctly
             && level.getFluidState(pos.below()).is(FluidTags.WATER)
             && level.getBlockState(pos.above()).is(Blocks.WATER);
     }
