--- a/net/minecraft/server/level/ChunkTrackingView.java
+++ b/net/minecraft/server/level/ChunkTrackingView.java
@@ -77,7 +_,7 @@
         return distanceSquared < radiusSquared;
     }
 
-    record Positioned(ChunkPos center, int viewDistance) implements ChunkTrackingView {
+    public record Positioned(ChunkPos center, int viewDistance) implements ChunkTrackingView {
         private int minX() {
             return this.center.x() - this.viewDistance - 1;
         }
@@ -95,7 +_,7 @@
         }
 
         @VisibleForTesting
-        boolean squareIntersects(ChunkTrackingView.Positioned other) {
+        protected boolean squareIntersects(ChunkTrackingView.Positioned other) {
             return this.minX() <= other.maxX() && this.maxX() >= other.minX() && this.minZ() <= other.maxZ() && this.maxZ() >= other.minZ();
         }
 
