--- a/net/minecraft/server/level/ChunkHolder.java
+++ b/net/minecraft/server/level/ChunkHolder.java
@@ -71,6 +_,18 @@
         this.changedBlocksPerSection = new ShortSet[p_142988_.getSectionsCount()];
     }
 
+    // CraftBukkit start
+    public LevelChunk getFullChunkNow() {
+        // Note: We use the oldTicketLevel for isLoaded checks.
+        if (!ChunkLevel.fullStatus(this.oldTicketLevel).isOrAfter(FullChunkStatus.FULL)) return null;
+        return this.getFullChunkNowUnchecked();
+    }
+
+    public LevelChunk getFullChunkNowUnchecked() {
+        return (LevelChunk) this.getChunkIfPresentUnchecked(ChunkStatus.FULL);
+    }
+    // CraftBukkit end
+
     public CompletableFuture<ChunkResult<LevelChunk>> getTickingChunkFuture() {
         return this.tickingChunkFuture;
     }
@@ -125,6 +_,7 @@
         LevelChunk levelchunk = this.getTickingChunk();
         if (levelchunk != null) {
             int i = this.levelHeightAccessor.getSectionIndex(p_140057_.getY());
+            if (i < 0 || i >= this.changedBlocksPerSection.length) return; // CraftBukkit - SPIGOT-6086, SPIGOT-6296
             if (this.changedBlocksPerSection[i] == null) {
                 this.hasChangedSections = true;
                 this.changedBlocksPerSection[i] = new ShortOpenHashSet();
@@ -262,6 +_,7 @@
         FullChunkStatus fullchunkstatus1 = ChunkLevel.fullStatus(this.ticketLevel);
         boolean flag = fullchunkstatus.isOrAfter(FullChunkStatus.FULL);
         boolean flag1 = fullchunkstatus1.isOrAfter(FullChunkStatus.FULL);
+
         this.wasAccessibleSinceLastSave |= flag1;
         if (!flag && flag1) {
             this.fullChunkFuture = p_143004_.prepareAccessibleChunk(this);
