--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -94,6 +_,7 @@
 import net.minecraft.world.level.chunk.storage.RegionStorageInfo;
 import net.minecraft.world.level.chunk.storage.SerializableChunkData;
 import net.minecraft.world.level.chunk.storage.SimpleRegionStorage;
+import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.entity.ChunkStatusUpdateListener;
 import net.minecraft.world.level.gamerules.GameRules;
 import net.minecraft.world.level.levelgen.NoiseBasedChunkGenerator;
@@ -150,7 +_,28 @@
     private final Queue<Runnable> unloadQueue = Queues.newConcurrentLinkedQueue();
     private final AtomicInteger activeChunkWrites = new AtomicInteger();
     public int serverViewDistance;
-    private final WorldGenContext worldGenContext;
+    public final WorldGenContext worldGenContext;
+
+    // CraftBukkit start - recursion-safe executor for Chunk loadCallback() and unloadCallback()
+    public final CallbackExecutor callbackExecutor = new CallbackExecutor();
+    public static final class CallbackExecutor implements java.util.concurrent.Executor, Runnable {
+
+        private final java.util.Queue<Runnable> queue = new java.util.ArrayDeque<>();
+
+        @Override
+        public void execute(Runnable runnable) {
+            queue.add(runnable);
+        }
+
+        @Override
+        public void run() {
+            Runnable task;
+            while ((task = queue.poll()) != null) {
+                task.run();
+            }
+        }
+    };
+    // CraftBukkit end
 
     public ChunkMap(
         ServerLevel level,
@@ -179,7 +_,13 @@
         this.level = level;
         RegistryAccess registryAccess = level.registryAccess();
         long levelSeed = level.getSeed();
-        if (generator instanceof NoiseBasedChunkGenerator noiseGenerator) {
+        // CraftBukkit start - SPIGOT-7051: It's a rigged game! Use delegate for random state creation, otherwise it is not so random.
+        ChunkGenerator randomGenerator = generator;
+        if (randomGenerator instanceof org.bukkit.craftbukkit.generator.CustomChunkGenerator customChunkGenerator) {
+            randomGenerator = customChunkGenerator.getDelegate();
+        }
+        if (randomGenerator instanceof NoiseBasedChunkGenerator noiseGenerator) {
+            // CraftBukkit end
             this.randomState = RandomState.create(noiseGenerator.generatorSettings().value(), registryAccess.lookupOrThrow(Registries.NOISE), levelSeed);
         } else {
             this.randomState = RandomState.create(NoiseGeneratorSettings.dummy(), registryAccess.lookupOrThrow(Registries.NOISE), levelSeed);
@@ -212,6 +_,47 @@
         this.chunksToEagerlySave.add(chunkPos.pack());
     }
 
+    // Paper start - Optional per player mob spawns
+    public void updatePlayerMobTypeMap(final Entity entity) {
+        if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) {
+            return;
+        }
+
+        final int index = entity.getType().getCategory().ordinal();
+        final ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> inRange =
+                this.level.moonrise$getNearbyPlayers().getPlayers(entity.chunkPosition(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+        if (inRange == null) {
+            return;
+        }
+
+        final ServerPlayer[] backingSet = inRange.getRawDataUnchecked();
+        for (int i = 0, len = inRange.size(); i < len; i++) {
+            ++(backingSet[i].mobCounts[index]);
+        }
+    }
+
+    // Paper start - per player mob count backoff
+    public void updateFailurePlayerMobTypeMap(int chunkX, int chunkZ, net.minecraft.world.entity.MobCategory mobCategory) {
+        if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) {
+            return;
+        }
+        int idx = mobCategory.ordinal();
+        final ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> inRange =
+                this.level.moonrise$getNearbyPlayers().getPlayersByChunk(chunkX, chunkZ, ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+        if (inRange == null) {
+            return;
+        }
+        final ServerPlayer[] backingSet = inRange.getRawDataUnchecked();
+        for (int i = 0, len = inRange.size(); i < len; i++) {
+            ++(backingSet[i].mobBackoffCounts[idx]);
+        }
+    }
+    // Paper end - per player mob count backoff
+    public int getMobCountNear(final ServerPlayer player, final net.minecraft.world.entity.MobCategory mobCategory) {
+        return player.mobCounts[mobCategory.ordinal()] + player.mobBackoffCounts[mobCategory.ordinal()]; // Paper - per player mob count backoff
+    }
+    // Paper end - Optional per player mob spawns
+
     protected ChunkGenerator generator() {
         return this.worldGenContext.generator();
     }
@@ -390,6 +_,7 @@
             this.modified = true;
         }
 
+        net.neoforged.neoforge.event.EventHooks.fireChunkTicketLevelUpdated(this.level, node, oldLevel, level, chunk);
         return chunk;
     }
 
@@ -524,9 +_,13 @@
 
                     this.save(chunk);
                     if (chunk instanceof LevelChunk levelChunk) {
+                        net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.ChunkEvent.Unload(levelChunk));
                         this.level.unload(levelChunk);
                     }
 
+                    net.neoforged.neoforge.common.CommonHooks.onChunkUnload(this.poiManager, chunk); // Neo: Must be called for all chunk unloading. Not just LevelChunks.
+                    this.chunkTypeCache.remove(chunk.getPos().pack()); // Neo: Prevent chunk type cache from permanently retaining data for unloaded chunks
+
                     this.lightEngine.updateChunkStatus(chunk.getPos());
                     this.lightEngine.tryScheduleUpdate();
                     this.nextChunkSaveTime.remove(chunk.getPos().pack());
@@ -566,6 +_,7 @@
                 Profiler.get().incrementCounter("chunkLoad");
                 if (chunkData.isPresent()) {
                     ChunkAccess chunk = chunkData.get().read(this.level, this.poiManager, this.storageInfo(), pos);
+                    net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.ChunkDataEvent.Load(chunk, chunkData.get()));
                     this.markPosition(pos, chunk.getPersistedStatus().getChunkType());
                     return chunk;
                 } else {
@@ -739,7 +_,7 @@
         }
     }
 
-    public boolean save(ChunkAccess chunk) {
+    private boolean save(ChunkAccess chunk) {
         this.poiManager.flush(chunk.getPos());
         if (!chunk.tryMarkSaved()) {
             return false;
@@ -762,6 +_,7 @@
             Profiler.get().incrementCounter("chunkSave");
             this.activeChunkWrites.incrementAndGet();
             SerializableChunkData data = SerializableChunkData.copyOf(this.level, chunk);
+            net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.ChunkDataEvent.Save(chunk, this.level, data));
             CompletableFuture<CompoundTag> encodedData = CompletableFuture.supplyAsync(data::write, Util.backgroundExecutor());
             this.write(pos, encodedData::join).handle((ignored, throwable) -> {
                 if (throwable != null) {
@@ -827,9 +_,11 @@
 
     private static void markChunkPendingToSend(ServerPlayer player, LevelChunk chunk) {
         player.connection.chunkSender.markChunkPendingToSend(chunk);
+        net.neoforged.neoforge.event.EventHooks.fireChunkWatch(player, chunk, player.level());
     }
 
     private static void dropChunk(ServerPlayer player, ChunkPos pos) {
+        net.neoforged.neoforge.event.EventHooks.fireChunkUnWatch(player, pos, player.level());
         player.connection.chunkSender.dropChunk(player, pos);
     }
 
@@ -914,16 +_,16 @@
         return this.read(pos).thenApplyAsync(chunkTag -> chunkTag.map(this::upgradeChunkTag), Util.backgroundExecutor().forName("upgradeChunk"));
     }
 
-    private CompoundTag upgradeChunkTag(CompoundTag tag) {
+    public CompoundTag upgradeChunkTag(CompoundTag tag) {
         return this.upgradeChunkTag(
             tag,
             -1,
-            getChunkDataFixContextTag(this.level.dimension(), this.generator().getTypeNameForDataFixer()),
+            getChunkDataFixContextTag(this.level.getTypeKey(), this.generator().getTypeNameForDataFixer()),
             SharedConstants.getCurrentVersion().dataVersion().version()
         );
     }
 
-    public static CompoundTag getChunkDataFixContextTag(ResourceKey<Level> dimension, Optional<Identifier> generatorIdentifier) {
+    public static CompoundTag getChunkDataFixContextTag(ResourceKey<LevelStem> dimension, Optional<Identifier> generatorIdentifier) {
         CompoundTag contextTag = new CompoundTag();
         contextTag.putString("dimension", dimension.identifier().toString());
         generatorIdentifier.ifPresent(identifier -> contextTag.putString("generator", identifier.toString()));
@@ -937,7 +_,7 @@
             ChunkHolder holder = this.visibleChunkMap.get(spawnCandidateChunks.nextLong());
             if (holder != null) {
                 LevelChunk chunk = holder.getTickingChunk();
-                if (chunk != null && this.anyPlayerCloseEnoughForSpawningInternal(holder.getPos())) {
+                if (chunk != null && (this.anyPlayerCloseEnoughForSpawningInternal(holder.getPos()) || ticketStorage.shouldForceNaturalSpawning(holder.getPos()))) {// Neo: Force natural spawning in other chunks
                     output.add(chunk);
                 }
             }
@@ -961,6 +_,19 @@
         return triState == TriState.DEFAULT ? this.anyPlayerCloseEnoughForSpawningInternal(pos) : triState.toBoolean(true);
     }
 
+    // TODO 1.21.11 unobfuscated - seems unused and requires an AT for ticketStorage
+//    void collectForceTickingChunks(List<LevelChunk> list) {
+//        for (long l : distanceManager.ticketStorage.getForceLoadedChunks()) {
+//            ChunkHolder holder = this.visibleChunkMap.get(l);
+//            if (holder != null) {
+//                var chunk = holder.getTickingChunk();
+//                if (chunk != null) {
+//                    list.add(chunk);
+//                }
+//            }
+//        }
+//    }
+
     public boolean anyPlayerCloseEnoughTo(BlockPos pos, int maxDistance) {
         Vec3 target = new Vec3(pos);
 
@@ -1090,6 +_,7 @@
                 this.playerMap.unIgnorePlayer(player);
             }
 
+            //PATCH 1.20.2: Figure out the firing of the watch and unwatch events when chunk tracking updates.
             this.updateChunkTracking(player);
         }
     }
@@ -1137,8 +_,8 @@
         return this.entityMap.containsKey(id);
     }
 
-    public void addEntity(Entity entity) {
-        if (!(entity instanceof EnderDragonPart)) {
+    protected void addEntity(Entity entity) {
+        if (!(entity instanceof net.neoforged.neoforge.entity.PartEntity)) {
             EntityType<?> type = entity.getType();
             int range = type.clientTrackingRange() * 16;
             if (range != 0) {
@@ -1296,7 +_,21 @@
         }
     }
 
-    public class DistanceManager extends net.minecraft.server.level.DistanceManager {
+    // Neo: Getter for players watching an entity.
+    public List<ServerPlayer> getPlayersWatching(Entity entity) {
+        var trackedEntity = entityMap.get(entity.getId());
+        if (trackedEntity != null) {
+            var ret = new java.util.ArrayList<ServerPlayer>(trackedEntity.seenBy.size());
+            for (var connection : trackedEntity.seenBy) {
+                ret.add(connection.getPlayer());
+            }
+            return java.util.Collections.unmodifiableList(ret);
+        } else {
+            return List.of();
+        }
+    }
+
+    private class DistanceManager extends net.minecraft.server.level.DistanceManager {
         protected DistanceManager(TicketStorage ticketStorage, Executor executor, Executor mainThreadExecutor) {
             super(ticketStorage, executor, mainThreadExecutor);
         }
@@ -1322,10 +_,11 @@
         private final Entity entity;
         private final int range;
         private SectionPos lastSectionPos;
-        public final Set<ServerPlayerConnection> seenBy = Sets.newIdentityHashSet();
+        public final Set<ServerPlayerConnection> seenBy;
 
         public TrackedEntity(Entity entity, int range, int updateInterval, boolean trackDelta) {
-            this.serverEntity = new ServerEntity(ChunkMap.this.level, entity, updateInterval, trackDelta, this);
+            this.seenBy = Sets.newIdentityHashSet();
+            this.serverEntity = new ServerEntity(ChunkMap.this.level, entity, updateInterval, trackDelta, this, seenBy); // CraftBukkit
             this.entity = entity;
             this.range = range;
             this.lastSectionPos = SectionPos.of(entity);
@@ -1365,6 +_,17 @@
             }
         }
 
+        // CraftBukkit start
+        @Override
+        public void sendToTrackingPlayersFilteredAndSelf(Packet<? super ClientGamePacketListener> packet, Predicate<ServerPlayer> predicate) {
+            this.sendToTrackingPlayersFiltered(packet, predicate);
+
+            if (this.entity instanceof ServerPlayer serverplayer) {
+                serverplayer.connection.send(packet);
+            }
+        }
+        // CraftBukkit end
+
         public void broadcastRemoved() {
             for (ServerPlayerConnection connection : this.seenBy) {
                 this.serverEntity.removePairing(connection.getPlayer());
@@ -1390,6 +_,11 @@
                 boolean visibleToPlayer = distanceSquared <= rangeSquared
                     && this.entity.broadcastToPlayer(player)
                     && ChunkMap.this.isChunkTracked(player, this.entity.chunkPosition().x(), this.entity.chunkPosition().z());
+                // CraftBukkit start - respect vanish API
+                if (!player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) {
+                    visibleToPlayer = false;
+                }
+                // CraftBukkit end
                 if (visibleToPlayer) {
                     if (this.seenBy.add(player.connection)) {
                         this.serverEntity.addPairing(player);
@@ -1427,5 +_,20 @@
                 this.updatePlayer(player);
             }
         }
+    }
+
+    /**
+     * Neo: PR #937
+     * This is for mainly pre-generation usage such as Neoforge's generate command.
+     * Use this to schedule chunk load tasks into ChunkTaskPriorityQueueSorter so a chunk is fully finished all of their tasks before scheduling more chunks to load.
+     * Reason for this is when scheduling a huge ton of Full Status chunk tasks to the server (to load chunks),
+     * you could cause the server to only process those loading tasks and never reach the two chunk tasks that are
+     * automatically scheduled to run after the chunk is loaded to Full. As a result of flooding the system with Full Status chunk tasks,
+     * the queue for the two kind of successor chunk tasks will grow and become a memory leak of lambdas and chunk references.
+     * Use this method to schedule tasks for loading chunks in your whenCompleteAsync method call so the tasks gets processed properly over time and not leak.
+     * See {@link net.neoforged.neoforge.server.command.generation.GenerationTask#enqueueChunks} as an example usage of this method.
+     */
+    public void scheduleOnMainThreadMailbox(Runnable runnable) {
+        mainThreadExecutor.schedule(runnable);
     }
 }
