--- a/net/minecraft/server/level/ClientInformation.java
+++ b/net/minecraft/server/level/ClientInformation.java
@@ -22,26 +_,26 @@
         this(
             input.readUtf(16),
             input.readByte(),
-            ChatVisiblity.STREAM_CODEC.decode(input),
+            input.readEnum(ChatVisiblity.class),
             input.readBoolean(),
             input.readUnsignedByte(),
-            HumanoidArm.STREAM_CODEC.decode(input),
-            input.readBoolean(),
-            input.readBoolean(),
-            ParticleStatus.STREAM_CODEC.decode(input)
+            input.readEnum(HumanoidArm.class),
+            input.readBoolean(),
+            input.readBoolean(),
+            input.readEnum(ParticleStatus.class)
         );
     }
 
     public void write(FriendlyByteBuf output) {
         output.writeUtf(this.language);
         output.writeByte(this.viewDistance);
-        ChatVisiblity.STREAM_CODEC.encode(output, this.chatVisibility);
+        output.writeEnum(this.chatVisibility);
         output.writeBoolean(this.chatColors);
         output.writeByte(this.modelCustomisation);
-        HumanoidArm.STREAM_CODEC.encode(output, this.mainHand);
+        output.writeEnum(this.mainHand);
         output.writeBoolean(this.textFilteringEnabled);
         output.writeBoolean(this.allowsListing);
-        ParticleStatus.STREAM_CODEC.encode(output, this.particleStatus);
+        output.writeEnum(this.particleStatus);
     }
 
     public static ClientInformation createDefault() {
