--- a/net/minecraft/server/level/DistanceManager.java
+++ b/net/minecraft/server/level/DistanceManager.java
@@ -17,6 +_,7 @@
 import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import it.unimi.dsi.fastutil.objects.ObjectSet;
 import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
+import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
@@ -31,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public abstract class DistanceManager {
+public abstract class DistanceManager implements ca.spottedleaf.moonrise.patches.chunk_system.level.chunk.ChunkSystemDistanceManager, ca.spottedleaf.moonrise.patches.chunk_tick_iteration.ChunkTickDistanceManager { // Paper - rewrite chunk system // Paper - chunk tick iteration optimisation
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int PLAYER_TICKET_LEVEL = ChunkLevel.byStatus(FullChunkStatus.ENTITY_TICKING);
     private final Long2ObjectMap<ObjectSet<ServerPlayer>> playersPerChunk = new Long2ObjectOpenHashMap<>();
@@ -72,6 +_,11 @@
         }
 
         if (!this.chunksToUpdateFutures.isEmpty()) {
+            // CraftBukkit start - SPIGOT-7780: Call chunk unload events before updateHighestAllowedStatus
+            for (ChunkHolder chunkholder : this.chunksToUpdateFutures) {
+                chunkholder.callEventIfUnloading(scheduler);
+            }
+            // CraftBukkit end
             for (ChunkHolder chunksToUpdateFuture : this.chunksToUpdateFutures) {
                 chunksToUpdateFuture.updateHighestAllowedStatus(scheduler);
             }
@@ -119,6 +_,7 @@
         ChunkPos chunk = pos.chunk();
         long chunkPos = chunk.pack();
         ObjectSet<ServerPlayer> chunkPlayers = this.playersPerChunk.get(chunkPos);
+        if (chunkPlayers == null) return; // CraftBukkit - SPIGOT-6208
         chunkPlayers.remove(player);
         if (chunkPlayers.isEmpty()) {
             this.playersPerChunk.remove(chunkPos);
