--- a/net/minecraft/server/level/DistanceManager.java
+++ b/net/minecraft/server/level/DistanceManager.java
@@ -31,29 +_,62 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public abstract class DistanceManager {
+public abstract class DistanceManager implements ca.spottedleaf.moonrise.patches.chunk_system.level.chunk.ChunkSystemDistanceManager, ca.spottedleaf.moonrise.patches.chunk_tick_iteration.ChunkTickDistanceManager { // Paper - rewrite chunk system // Paper - chunk tick iteration optimisation
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int PLAYER_TICKET_LEVEL = ChunkLevel.byStatus(FullChunkStatus.ENTITY_TICKING);
     private final Long2ObjectMap<ObjectSet<ServerPlayer>> playersPerChunk = new Long2ObjectOpenHashMap<>();
-    private final LoadingChunkTracker loadingChunkTracker;
-    private final SimulationChunkTracker simulationChunkTracker;
+    // Paper - rewrite chunk system
     public final TicketStorage ticketStorage;
-    private final DistanceManager.FixedPlayerDistanceChunkTracker naturalSpawnChunkCounter = new DistanceManager.FixedPlayerDistanceChunkTracker(8);
-    private final DistanceManager.PlayerTicketTracker playerTicketManager = new DistanceManager.PlayerTicketTracker(32);
-    protected final Set<ChunkHolder> chunksToUpdateFutures = new ReferenceOpenHashSet<>();
-    private final ThrottlingChunkTaskDispatcher ticketDispatcher;
-    private final LongSet ticketsToRelease = new LongOpenHashSet();
-    private final Executor mainThreadExecutor;
-    public int simulationDistance = 10;
+    // Paper - chunk tick iteration optimisation
+    // Paper - rewrite chunk system
 
-    protected DistanceManager(TicketStorage ticketStorage, Executor executor, Executor mainThreadExecutor) {
+    protected DistanceManager(final TicketStorage ticketStorage, final Executor executor, final Executor mainThreadExecutor) {
         this.ticketStorage = ticketStorage;
-        this.loadingChunkTracker = new LoadingChunkTracker(this, ticketStorage);
-        this.simulationChunkTracker = new SimulationChunkTracker(ticketStorage);
+        // Paper - rewrite chunk system
         TaskScheduler<Runnable> mainThreadTaskScheduler = TaskScheduler.wrapExecutor("player ticket throttler", mainThreadExecutor);
-        this.ticketDispatcher = new ThrottlingChunkTaskDispatcher(mainThreadTaskScheduler, executor, 4);
-        this.mainThreadExecutor = mainThreadExecutor;
-    }
+        this.ticketStorage.moonrise$setChunkMap(this.moonrise$getChunkMap()); // Paper - rewrite chunk system
+    }
+
+    // Paper start - rewrite chunk system
+    @Override
+    public final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkHolderManager moonrise$getChunkHolderManager() {
+        return ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.moonrise$getChunkMap().level).moonrise$getChunkTaskScheduler().chunkHolderManager;
+    }
+    // Paper end - rewrite chunk system
+    // Paper start - chunk tick iteration optimisation
+    private final ca.spottedleaf.moonrise.common.misc.PositionCountingAreaMap<ServerPlayer> spawnChunkTracker = new ca.spottedleaf.moonrise.common.misc.PositionCountingAreaMap<>();
+    // Note: Cannot do narrow tracking on Paper due to custom spawn range
+
+    @Override
+    public final void moonrise$addPlayer(final ServerPlayer player, final SectionPos pos) {
+        this.spawnChunkTracker.add(player, pos.x(), pos.z(), ca.spottedleaf.moonrise.patches.chunk_tick_iteration.ChunkTickConstants.PLAYER_SPAWN_TRACK_RANGE);
+        // Note: Cannot do narrow tracking on Paper due to custom spawn range
+    }
+
+    @Override
+    public final void moonrise$removePlayer(final ServerPlayer player, final SectionPos pos) {
+        this.spawnChunkTracker.remove(player);
+        // Note: Cannot do narrow tracking on Paper due to custom spawn range
+    }
+
+    @Override
+    public final void moonrise$updatePlayer(final ServerPlayer player,
+                                            final SectionPos oldPos, final SectionPos newPos,
+                                            final boolean oldIgnore, final boolean newIgnore) {
+        if (newIgnore) {
+            this.spawnChunkTracker.remove(player);
+            // Note: Cannot do narrow tracking on Paper due to custom spawn range
+        } else {
+            this.spawnChunkTracker.addOrUpdate(player, newPos.x(), newPos.z(), ca.spottedleaf.moonrise.patches.chunk_tick_iteration.ChunkTickConstants.PLAYER_SPAWN_TRACK_RANGE);
+            // Note: Cannot do narrow tracking on Paper due to custom spawn range
+        }
+    }
+
+    @Override
+    public final boolean moonrise$hasAnyNearbyNarrow(final int chunkX, final int chunkZ) {
+        throw new UnsupportedOperationException(); // Note: Cannot do narrow tracking on Paper due to custom spawn range
+    }
+    // Paper end - chunk tick iteration optimisation
 
     protected abstract boolean isChunkToRemove(final long node);
 
@@ -61,132 +_,100 @@
 
     protected abstract @Nullable ChunkHolder updateChunkScheduling(final long node, final int level, final @Nullable ChunkHolder chunk, final int oldLevel);
 
-    public boolean runAllUpdates(ChunkMap scheduler) {
-        this.naturalSpawnChunkCounter.runAllUpdates();
-        this.simulationChunkTracker.runAllUpdates();
-        this.playerTicketManager.runAllUpdates();
-        int updates = Integer.MAX_VALUE - this.loadingChunkTracker.runDistanceUpdates(Integer.MAX_VALUE);
-        boolean updated = updates != 0;
-        if (updated && SharedConstants.DEBUG_VERBOSE_SERVER_EVENTS) {
-            LOGGER.debug("DMU {}", updates);
-        }
-
-        if (!this.chunksToUpdateFutures.isEmpty()) {
-            for (ChunkHolder chunksToUpdateFuture : this.chunksToUpdateFutures) {
-                chunksToUpdateFuture.updateHighestAllowedStatus(scheduler);
-            }
-
-            for (ChunkHolder chunkHolder : this.chunksToUpdateFutures) {
-                chunkHolder.updateFutures(scheduler, this.mainThreadExecutor);
-            }
-
-            this.chunksToUpdateFutures.clear();
-            return true;
-        } else {
-            if (!this.ticketsToRelease.isEmpty()) {
-                LongIterator iterator = this.ticketsToRelease.iterator();
-
-                while (iterator.hasNext()) {
-                    long pos = iterator.nextLong();
-                    if (this.ticketStorage.getTickets(pos).stream().anyMatch(t -> t.getType() == TicketType.PLAYER_LOADING)) {
-                        ChunkHolder chunk = scheduler.getUpdatingChunkIfPresent(pos);
-                        if (chunk == null) {
-                            throw new IllegalStateException();
-                        }
-
-                        CompletableFuture<ChunkResult<LevelChunk>> future = chunk.getEntityTickingChunkFuture();
-                        future.thenAccept(c -> this.mainThreadExecutor.execute(() -> this.ticketDispatcher.release(pos, () -> {}, false)));
-                    }
-                }
-
-                this.ticketsToRelease.clear();
-            }
-
-            return updated;
-        }
+    public boolean runAllUpdates(final ChunkMap scheduler) {
+        return this.moonrise$getChunkHolderManager().processTicketUpdates(); // Paper - rewrite chunk system
     }
 
-    public void addPlayer(SectionPos pos, ServerPlayer player) {
+    public void addPlayer(final SectionPos pos, final ServerPlayer player) {
         ChunkPos chunk = pos.chunk();
         long chunkPos = chunk.pack();
         this.playersPerChunk.computeIfAbsent(chunkPos, k -> new ObjectOpenHashSet<>()).add(player);
-        this.naturalSpawnChunkCounter.update(chunkPos, 0, true);
-        this.playerTicketManager.update(chunkPos, 0, true);
-        this.ticketStorage.addTicket(new Ticket(TicketType.PLAYER_SIMULATION, this.getPlayerTicketLevel()), chunk);
+        // Paper - chunk tick iteration optimisation
+        // Paper - rewrite chunk system
     }
 
-    public void removePlayer(SectionPos pos, ServerPlayer player) {
+    public void removePlayer(final SectionPos pos, final ServerPlayer player) {
         ChunkPos chunk = pos.chunk();
         long chunkPos = chunk.pack();
         ObjectSet<ServerPlayer> chunkPlayers = this.playersPerChunk.get(chunkPos);
-        chunkPlayers.remove(player);
-        if (chunkPlayers.isEmpty()) {
+        // Paper start - some state corruption happens here, don't crash, clean up gracefully
+        if (chunkPlayers != null) chunkPlayers.remove(player);
+        if (chunkPlayers == null || chunkPlayers.isEmpty()) {
+            // Paper end - some state corruption happens here, don't crash, clean up gracefully
             this.playersPerChunk.remove(chunkPos);
-            this.naturalSpawnChunkCounter.update(chunkPos, Integer.MAX_VALUE, false);
-            this.playerTicketManager.update(chunkPos, Integer.MAX_VALUE, false);
-            this.ticketStorage.removeTicket(new Ticket(TicketType.PLAYER_SIMULATION, this.getPlayerTicketLevel()), chunk);
+            // Paper - chunk tick iteration optimisation
+            // Paper - rewrite chunk system
         }
     }
 
     private int getPlayerTicketLevel() {
-        return Math.max(0, ChunkLevel.byStatus(FullChunkStatus.ENTITY_TICKING) - this.simulationDistance);
-    }
-
-    public boolean inEntityTickingRange(long key) {
-        return ChunkLevel.isEntityTicking(this.simulationChunkTracker.getLevel(key));
-    }
-
-    public boolean inBlockTickingRange(long key) {
-        return ChunkLevel.isBlockTicking(this.simulationChunkTracker.getLevel(key));
-    }
-
-    public int getChunkLevel(long key, boolean simulation) {
-        return simulation ? this.simulationChunkTracker.getLevel(key) : this.loadingChunkTracker.getLevel(key);
-    }
-
-    protected void updatePlayerTickets(int viewDistance) {
-        this.playerTicketManager.updateViewDistance(viewDistance);
-    }
-
-    public void updateSimulationDistance(int newDistance) {
-        if (newDistance != this.simulationDistance) {
-            this.simulationDistance = newDistance;
-            this.ticketStorage.replaceTicketLevelOfType(this.getPlayerTicketLevel(), TicketType.PLAYER_SIMULATION);
-        }
+        throw new UnsupportedOperationException(); // Paper - rewrite chunk system
+    }
+
+    public boolean inEntityTickingRange(final long key) {
+        // Paper start - rewrite chunk system
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = this.moonrise$getChunkHolderManager().getChunkHolder(key);
+        return chunkHolder != null && chunkHolder.isEntityTickingReady();
+        // Paper end - rewrite chunk system
+    }
+
+    public boolean inBlockTickingRange(final long key) {
+        // Paper start - rewrite chunk system
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = this.moonrise$getChunkHolderManager().getChunkHolder(key);
+        return chunkHolder != null && chunkHolder.isTickingReady();
+        // Paper end - rewrite chunk system
+    }
+
+    public int getChunkLevel(final long key, final boolean simulation) {
+        // Paper start - rewrite chunk system
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = this.moonrise$getChunkHolderManager().getChunkHolder(key);
+        return chunkHolder == null ? ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkHolderManager.MAX_TICKET_LEVEL + 1 : chunkHolder.getTicketLevel();
+        // Paper end - rewrite chunk system
+    }
+
+    protected void updatePlayerTickets(final int viewDistance) {
+        this.moonrise$getChunkMap().setServerViewDistance(viewDistance); // Paper - rewrite chunk system
+    }
+
+    public void updateSimulationDistance(final int newDistance) {
+        // Paper start - rewrite chunk system
+        // note: vanilla does not clamp to 0, but we do simply because we need a min of 0
+        final int clamped = net.minecraft.util.Mth.clamp(newDistance, 0, ca.spottedleaf.moonrise.common.util.MoonriseConstants.MAX_VIEW_DISTANCE);
+
+        ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.moonrise$getChunkMap().level).moonrise$getPlayerChunkLoader().setTickDistance(clamped);
+        // Paper end - rewrite chunk system
     }
 
     public int getNaturalSpawnChunkCount() {
-        this.naturalSpawnChunkCounter.runAllUpdates();
-        return this.naturalSpawnChunkCounter.chunks.size();
-    }
-
-    public TriState hasPlayersNearby(long pos) {
-        this.naturalSpawnChunkCounter.runAllUpdates();
-        int distance = this.naturalSpawnChunkCounter.getLevel(pos);
-        if (distance <= NaturalSpawner.INSCRIBED_SQUARE_SPAWN_DISTANCE_CHUNK) {
-            return TriState.TRUE;
-        } else {
-            return distance > 8 ? TriState.FALSE : TriState.DEFAULT;
-        }
-    }
-
-    public void forEachEntityTickingChunk(LongConsumer consumer) {
-        for (Entry entry : Long2ByteMaps.fastIterable(this.simulationChunkTracker.chunks)) {
-            byte level = entry.getByteValue();
-            long key = entry.getLongKey();
-            if (ChunkLevel.isEntityTicking(level)) {
-                consumer.accept(key);
-            }
-        }
+        return this.spawnChunkTracker.getTotalPositions(); // Paper - chunk tick iteration optimisation
+    }
+
+    public TriState hasPlayersNearby(final long pos) {
+        // Note: Cannot do narrow tracking on Paper due to custom spawn range // Paper - chunk tick iteration optimisation
+        return this.spawnChunkTracker.hasObjectsNear(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkX(pos), ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkZ(pos)) ? net.minecraft.util.TriState.DEFAULT : net.minecraft.util.TriState.FALSE; // Paper - chunk tick iteration optimisation
+    }
+
+    public void forEachEntityTickingChunk(final LongConsumer consumer) {
+        // Paper start - rewrite chunk system
+        final ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.world.level.chunk.LevelChunk> chunks = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.moonrise$getChunkMap().level).moonrise$getEntityTickingChunks();
+        final LevelChunk[] raw = chunks.getRawDataUnchecked();
+        final int size = chunks.size();
+
+        java.util.Objects.checkFromToIndex(0, size, raw.length);
+        for (int i = 0; i < size; ++i) {
+            final LevelChunk chunk = raw[i];
+
+            consumer.accept(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunk.getPos()));
+        }
+        // Paper end - rewrite chunk system
     }
 
     public LongIterator getSpawnCandidateChunks() {
-        this.naturalSpawnChunkCounter.runAllUpdates();
-        return this.naturalSpawnChunkCounter.chunks.keySet().iterator();
+        return this.spawnChunkTracker.getPositions().iterator(); // Paper - chunk tick iteration optimisation
     }
 
     public String getDebugStatus() {
-        return this.ticketDispatcher.getDebugStatus();
+        return "N/A"; // Paper - rewrite chunk system
     }
 
     public boolean hasTickets() {
@@ -197,19 +_,19 @@
         protected final Long2ByteMap chunks = new Long2ByteOpenHashMap();
         protected final int maxDistance;
 
-        protected FixedPlayerDistanceChunkTracker(int maxDistance) {
+        protected FixedPlayerDistanceChunkTracker(final int maxDistance) {
             super(maxDistance + 2, 16, 256);
             this.maxDistance = maxDistance;
             this.chunks.defaultReturnValue((byte)(maxDistance + 2));
         }
 
         @Override
-        protected int getLevel(long node) {
+        protected int getLevel(final long node) {
             return this.chunks.get(node);
         }
 
         @Override
-        protected void setLevel(long node, int level) {
+        protected void setLevel(final long node, final int level) {
             byte oldLevel;
             if (level > this.maxDistance) {
                 oldLevel = this.chunks.remove(node);
@@ -220,15 +_,15 @@
             this.onLevelChange(node, oldLevel, level);
         }
 
-        protected void onLevelChange(long node, int oldLevel, int level) {
+        protected void onLevelChange(final long node, final int oldLevel, final int level) {
         }
 
         @Override
-        protected int getLevelFromSource(long to) {
+        protected int getLevelFromSource(final long to) {
             return this.havePlayer(to) ? 0 : Integer.MAX_VALUE;
         }
 
-        private boolean havePlayer(long chunkPos) {
+        private boolean havePlayer(final long chunkPos) {
             ObjectSet<ServerPlayer> players = DistanceManager.this.playersPerChunk.get(chunkPos);
             return players != null && !players.isEmpty();
         }
@@ -238,23 +_,24 @@
         }
     }
 
+/*  // Paper - rewrite chunk system
     private class PlayerTicketTracker extends DistanceManager.FixedPlayerDistanceChunkTracker {
         private int viewDistance;
         private final Long2IntMap queueLevels = Long2IntMaps.synchronize(new Long2IntOpenHashMap());
         private final LongSet toUpdate = new LongOpenHashSet();
 
-        protected PlayerTicketTracker(int maxDistance) {
+        protected PlayerTicketTracker(final int maxDistance) {
             super(maxDistance);
             this.viewDistance = 0;
             this.queueLevels.defaultReturnValue(maxDistance + 2);
         }
 
         @Override
-        protected void onLevelChange(long node, int oldLevel, int level) {
+        protected void onLevelChange(final long node, final int oldLevel, final int level) {
             this.toUpdate.add(node);
         }
 
-        public void updateViewDistance(int viewDistance) {
+        public void updateViewDistance(final int viewDistance) {
             for (Entry entry : this.chunks.long2ByteEntrySet()) {
                 byte level = entry.getByteValue();
                 long key = entry.getLongKey();
@@ -264,7 +_,7 @@
             this.viewDistance = viewDistance;
         }
 
-        private void onLevelChange(long key, int level, boolean saw, boolean sees) {
+        private void onLevelChange(final long key, final int level, final boolean saw, final boolean sees) {
             if (saw != sees) {
                 Ticket ticket = new Ticket(TicketType.PLAYER_LOADING, DistanceManager.PLAYER_TICKET_LEVEL);
                 if (sees) {
@@ -313,8 +_,8 @@
             }
         }
 
-        private boolean haveTicketFor(int level) {
+        private boolean haveTicketFor(final int level) {
             return level <= this.viewDistance;
         }
-    }
+    }*/  // Paper - rewrite chunk system
 }
