--- a/net/minecraft/server/level/DistanceManager.java
+++ b/net/minecraft/server/level/DistanceManager.java
@@ -17,6 +_,7 @@
 import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
 import it.unimi.dsi.fastutil.objects.ObjectSet;
 import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
+import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Executor;
@@ -72,6 +_,11 @@
         }
 
         if (!this.chunksToUpdateFutures.isEmpty()) {
+            // CraftBukkit start - SPIGOT-7780: Call chunk unload events before updateHighestAllowedStatus
+            for (ChunkHolder chunkholder : this.chunksToUpdateFutures) {
+                chunkholder.callEventIfUnloading(scheduler);
+            }
+            // CraftBukkit end
             for (ChunkHolder chunksToUpdateFuture : this.chunksToUpdateFutures) {
                 chunksToUpdateFuture.updateHighestAllowedStatus(scheduler);
             }
@@ -119,6 +_,7 @@
         ChunkPos chunk = pos.chunk();
         long chunkPos = chunk.pack();
         ObjectSet<ServerPlayer> chunkPlayers = this.playersPerChunk.get(chunkPos);
+        if (chunkPlayers == null) return; // CraftBukkit - SPIGOT-6208
         chunkPlayers.remove(player);
         if (chunkPlayers.isEmpty()) {
             this.playersPerChunk.remove(chunkPos);
