--- a/net/minecraft/server/level/LoadingChunkTracker.java
+++ b/net/minecraft/server/level/LoadingChunkTracker.java
@@ -38,7 +_,7 @@
         if (oldLevel != level) {
             chunk = this.distanceManager.updateChunkScheduling(node, level, chunk, oldLevel);
             if (chunk != null) {
-                this.distanceManager.chunksToUpdateFutures.add(chunk);
+                // Paper - rewrite chunk system
             }
         }
     }
