--- a/net/minecraft/server/level/PlayerRespawnLogic.java
+++ b/net/minecraft/server/level/PlayerRespawnLogic.java
@@ -13,7 +_,7 @@
 
 public class PlayerRespawnLogic {
     @Nullable
-    public static BlockPos getOverworldRespawnPos(ServerLevel p_183929_, int p_183930_, int p_183931_) {
+    protected static BlockPos getOverworldRespawnPos(ServerLevel p_183929_, int p_183930_, int p_183931_) {
         boolean flag = p_183929_.dimensionType().hasCeiling();
         LevelChunk levelchunk = p_183929_.getChunk(SectionPos.blockToSectionCoord(p_183930_), SectionPos.blockToSectionCoord(p_183931_));
         int i = flag
