--- a/net/minecraft/server/level/ServerBossEvent.java
+++ b/net/minecraft/server/level/ServerBossEvent.java
@@ -15,7 +_,7 @@
 public class ServerBossEvent extends BossEvent {
     private final Set<ServerPlayer> players = Sets.newHashSet();
     private final Set<ServerPlayer> unmodifiablePlayers = Collections.unmodifiableSet(this.players);
-    private boolean visible = true;
+    public boolean visible = true;
 
     public ServerBossEvent(UUID id, Component name, BossEvent.BossBarColor color, BossEvent.BossBarOverlay overlay) {
         super(id, name, color, overlay);
