--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -15,6 +_,7 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.BooleanSupplier;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
@@ -63,7 +_,7 @@
     public final ServerChunkCache.MainThreadExecutor mainThreadProcessor;
     public final ChunkMap chunkMap;
     private final SavedDataStorage savedDataStorage;
-    private final TicketStorage ticketStorage;
+    public final TicketStorage ticketStorage;
     private long lastInhabitedUpdate;
     public boolean spawnEnemies = true;
     private static final int CACHE_SIZE = 4;
@@ -99,7 +_,7 @@
             LOGGER.error("Failed to create dimension data storage directory", (Throwable)var14);
         }
 
-        this.savedDataStorage = new SavedDataStorage(dataFolder, fixerUpper, level.registryAccess());
+        this.savedDataStorage = new SavedDataStorage(level, dataFolder, fixerUpper, level.registryAccess());
         this.ticketStorage = this.savedDataStorage.computeIfAbsent(TicketStorage.TYPE);
         this.chunkMap = new ChunkMap(
             level,
@@ -122,6 +_,16 @@
         this.clearCache();
     }
 
+    // CraftBukkit start - properly implement isChunkLoaded
+    public boolean isChunkLoaded(int chunkX, int chunkZ) {
+        ChunkHolder chunk = this.chunkMap.getUpdatingChunkIfPresent(ChunkPos.pack(chunkX, chunkZ));
+        if (chunk == null) {
+            return false;
+        }
+        return chunk.getFullChunkNow() != null;
+    }
+    // CraftBukkit end
+
     public ThreadedLevelLightEngine getLightEngine() {
         return this.lightEngine;
     }
@@ -154,12 +_,16 @@
             for (int i = 0; i < 4; i++) {
                 if (pos == this.lastChunkPos[i] && targetStatus == this.lastChunkStatus[i]) {
                     ChunkAccess chunkAccess = this.lastChunk[i];
-                    if (chunkAccess != null || !loadOrGenerate) {
+                    if (chunkAccess != null) { // CraftBukkit - the chunk can become accessible in the meantime TODO for non-null chunks it might also make sense to check that the chunk's state hasn't changed in the meantime
                         return chunkAccess;
                     }
                 }
             }
 
+            ChunkHolder chunkholder = this.getVisibleChunkIfPresent(pos);
+            if (chunkholder != null && chunkholder.currentlyLoading != null)
+                return chunkholder.currentlyLoading; // Neo: If the requested chunk is loading, bypass the future chain to prevent a deadlock.
+
             profiler.incrementCounter("getChunkCacheMiss");
             CompletableFuture<ChunkResult<ChunkAccess>> serverFuture = this.getChunkFutureMainThread(x, z, targetStatus, loadOrGenerate);
             this.mainThreadProcessor.managedBlock(serverFuture::isDone);
@@ -193,6 +_,7 @@
             if (chunkHolder == null) {
                 return null;
             } else {
+                if (chunkHolder.currentlyLoading != null) return chunkHolder.currentlyLoading; // Forge: If the requested chunk is loading, bypass the future chain to prevent a deadlock.
                 ChunkAccess chunk = chunkHolder.getChunkIfPresent(ChunkStatus.FULL);
                 if (chunk != null) {
                     this.storeInCache(pos, chunk, ChunkStatus.FULL);
@@ -233,7 +_,15 @@
         long key = pos.pack();
         int targetTicketLevel = ChunkLevel.byStatus(targetStatus);
         ChunkHolder chunkHolder = this.getVisibleChunkIfPresent(key);
-        if (loadOrGenerate) {
+        // CraftBukkit start - don't add new ticket for currently unloading chunk
+        boolean currentlyUnloading = false;
+        if (chunkHolder != null) {
+            FullChunkStatus oldChunkState = ChunkLevel.fullStatus(chunkHolder.oldTicketLevel);
+            FullChunkStatus currentChunkState = ChunkLevel.fullStatus(chunkHolder.getTicketLevel());
+            currentlyUnloading = (oldChunkState.isOrAfter(FullChunkStatus.FULL) && !currentChunkState.isOrAfter(FullChunkStatus.FULL));
+        }
+        if (loadOrGenerate && !currentlyUnloading) {
+            // CraftBukkit end
             this.addTicket(new Ticket(TicketType.UNKNOWN, targetTicketLevel), pos);
             if (this.chunkAbsent(chunkHolder, targetTicketLevel)) {
                 ProfilerFiller profiler = Profiler.get();
@@ -253,7 +_,7 @@
     }
 
     private boolean chunkAbsent(@Nullable ChunkHolder chunkHolder, int targetTicketLevel) {
-        return chunkHolder == null || chunkHolder.getTicketLevel() > targetTicketLevel;
+        return chunkHolder == null || chunkHolder.oldTicketLevel > targetTicketLevel; // CraftBukkit using oldTicketLevel for isLoaded checks
     }
 
     @Override
@@ -304,19 +_,41 @@
         this.chunkMap.saveAllChunks(flushStorage);
     }
 
+    public AtomicBoolean close_save = new AtomicBoolean(true);
     @Override
     public void close() throws IOException {
-        this.save(true);
+        if (this.close_save.getAndSet(true)) {
+            this.save(true);
+        }
         this.savedDataStorage.close();
         this.lightEngine.close();
         this.chunkMap.close();
     }
 
+    public void close(boolean save) throws IOException {
+        this.close_save.set(save);
+        this.close();
+    }
+
+    // CraftBukkit start - modelled on below
+    public void purgeUnload() {
+        ProfilerFiller profilerfiller = Profiler.get();
+
+        profilerfiller.push("purge");
+        this.ticketStorage.purgeStaleTickets(this.chunkMap);
+        this.runDistanceManagerUpdates();
+        profilerfiller.popPush("unload");
+        this.chunkMap.tick(() -> true);
+        profilerfiller.pop();
+        this.clearCache();
+    }
+    // CraftBukkit end
+
     @Override
     public void tick(BooleanSupplier haveTime, boolean tickChunks) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("purge");
-        if (this.level.tickRateManager().runsNormally() || !tickChunks) {
+        if (this.level.tickRateManager().runsNormally() || !tickChunks || this.level.spigotConfig.unloadFrozenChunks) { // Spigot
             this.ticketStorage.purgeStaleTickets(this.chunkMap);
         }
 
@@ -372,12 +_,14 @@
             chunkCount, this.level.getAllEntities(), this::getFullChunk, new LocalMobCapCalculator(this.chunkMap)
         );
         this.lastSpawnState = spawnCookie;
-        boolean doMobSpawning = this.level.getGameRules().get(GameRules.SPAWN_MOBS);
+        boolean doMobSpawning = this.level.getGameRules().get(GameRules.SPAWN_MOBS) && !this.level.players().isEmpty(); // CraftBukkit
         int tickSpeed = this.level.getGameRules().get(GameRules.RANDOM_TICK_SPEED);
         List<MobCategory> spawningCategories;
         if (doMobSpawning) {
-            boolean spawnPersistent = this.level.getGameTime() % 400L == 0L;
+            boolean spawnPersistent = this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) != 0L && this.level.getLevelData().getGameTime() % this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) == 0L; // CraftBukkit
+            NaturalSpawner.serverlevel = this.level; // CraftBukkit
             spawningCategories = NaturalSpawner.getFilteredSpawningCategories(spawnCookie, true, this.spawnEnemies, spawnPersistent);
+            NaturalSpawner.serverlevel = null;
         } else {
             spawningCategories = List.of();
         }
@@ -617,13 +_,19 @@
         }
 
         @Override
-        protected boolean pollTask() {
+        // CraftBukkit start - process pending Chunk loadCallback() and unloadCallback() after each run task
+        public boolean pollTask() {
+        try {
             if (ServerChunkCache.this.runDistanceManagerUpdates()) {
                 return true;
             } else {
                 ServerChunkCache.this.lightEngine.tryScheduleUpdate();
                 return super.pollTask();
             }
+        } finally {
+            chunkMap.callbackExecutor.run();
+        }
+        // CraftBukkit end
         }
     }
 }
