--- a/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -7,6 +_,7 @@
 import net.minecraft.advancements.triggers.CriteriaTriggers;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
+import net.minecraft.network.protocol.game.ClientboundBlockChangedAckPacket;
 import net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket;
 import net.minecraft.network.protocol.game.ClientboundPlayerInfoUpdatePacket;
 import net.minecraft.network.protocol.game.ServerboundPlayerActionPacket;
@@ -22,12 +_,20 @@
 import net.minecraft.world.level.GameType;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.DoorBlock;
 import net.minecraft.world.level.block.GameMasterBlock;
+import net.minecraft.world.level.block.TrapDoorBlock;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.properties.DoubleBlockHalf;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.Event;
+import org.bukkit.event.block.Action;
+import org.bukkit.event.player.PlayerGameModeChangeEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
@@ -56,7 +_,13 @@
         if (gameModeForPlayer == this.gameModeForPlayer) {
             return false;
         }
-
+        // CraftBukkit start
+        PlayerGameModeChangeEvent event = new PlayerGameModeChangeEvent(player.getBukkitEntity(), org.bukkit.GameMode.getByValue(gameModeForPlayer.getId()));
+        org.bukkit.Bukkit.getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            return false;
+        }
+        // CraftBukkit end
         Abilities abilities = this.player.getAbilities();
         this.setGameModeForPlayer(gameModeForPlayer, this.gameModeForPlayer);
         if (abilities.flying && gameModeForPlayer != GameType.SPECTATOR && this.isInRangeOfGround()) {
@@ -80,7 +_,10 @@
         this.previousGameModeForPlayer = previousGameModeForPlayer;
         this.gameModeForPlayer = gameModeForPlayer;
         Abilities abilities = this.player.getAbilities();
+        // Neo: preserve flying state, removed on tick if Attribute or ability no longer applies
+        boolean wasFlying = abilities.flying;
         gameModeForPlayer.updatePlayerAbilities(abilities);
+        abilities.flying = wasFlying || abilities.flying;
     }
 
     private boolean isInRangeOfGround() {
@@ -105,7 +_,7 @@
     }
 
     public void tick() {
-        this.gameTicks++;
+        this.gameTicks = net.minecraft.server.MinecraftServer.currentTick; // CraftBukkit
         if (this.hasDelayedDestroy) {
             BlockState blockState = this.level.getBlockState(this.delayedDestroyPos);
             if (blockState.isAir()) {
@@ -148,6 +_,10 @@
     }
 
     public void handleBlockBreakAction(BlockPos pos, ServerboundPlayerActionPacket.Action action, Direction direction, int maxY, int sequence) {
+        net.neoforged.neoforge.event.entity.player.PlayerInteractEvent.LeftClickBlock event = net.neoforged.neoforge.common.CommonHooks.onLeftClickBlock(player, pos, direction, action);
+        if (event.isCanceled()) {
+            return;
+        }
         if (!this.player.isWithinBlockInteractionRange(pos, 1.0)) {
             this.debugLogging(pos, false, sequence, "too far");
         } else if (pos.getY() > maxY) {
@@ -162,16 +_,50 @@
                 }
 
                 if (!this.level.mayInteract(this.player, pos)) {
+                    // CraftBukkit start - fire PlayerInteractEvent
+                    CraftEventFactory.callPlayerInteractEvent(this.player, Action.LEFT_CLICK_BLOCK, pos, direction, this.player.getInventory().getSelectedItem(), InteractionHand.MAIN_HAND);
                     this.player.connection.send(new ClientboundBlockUpdatePacket(pos, this.level.getBlockState(pos)));
                     this.debugLogging(pos, false, sequence, "may not interact");
-                    return;
-                }
+                    // SPIGOT-8042: Need to ACK above changes first
+                    this.player.connection.send(new ClientboundBlockChangedAckPacket(sequence));
+                    // Update any tile entity data for this block
+                    BlockEntity blockentity = level.getBlockEntity(pos);
+                    if (blockentity != null) {
+                        this.player.connection.send(blockentity.getUpdatePacket());
+                    }
+                    // CraftBukkit end
+                    return;
+                }
+
+                // CraftBukkit start
+                PlayerInteractEvent bukkit_event = CraftEventFactory.callPlayerInteractEvent(this.player, Action.LEFT_CLICK_BLOCK, pos, direction, this.player.getInventory().getSelectedItem(), InteractionHand.MAIN_HAND);
+                if (bukkit_event.isCancelled()) {
+                    // Let the client know the block still exists
+                    this.player.connection.send(new ClientboundBlockUpdatePacket(this.level, pos));
+                    // SPIGOT-8042: Need to ACK above changes first
+                    this.player.connection.send(new ClientboundBlockChangedAckPacket(sequence));
+                    // Update any tile entity data for this block
+                    BlockEntity blockentity = this.level.getBlockEntity(pos);
+                    if (blockentity != null) {
+                        this.player.connection.send(blockentity.getUpdatePacket());
+                    }
+                    return;
+                }
+                // CraftBukkit end
 
                 if (this.player.getAbilities().instabuild) {
                     this.destroyAndAck(pos, sequence, "creative destroy");
                     return;
                 }
 
+                // Spigot start - handle debug stick left click for non-creative
+                if (this.player.getMainHandItem().is(net.minecraft.world.item.Items.DEBUG_STICK)
+                        && ((net.minecraft.world.item.DebugStickItem) net.minecraft.world.item.Items.DEBUG_STICK).handleInteraction(this.player, this.level.getBlockState(pos), this.level, pos, false, this.player.getMainHandItem())) {
+                    this.player.connection.send(new ClientboundBlockUpdatePacket(this.level, pos));
+                    return;
+                }
+                // Spigot end
+
                 if (this.player.blockActionRestricted(this.level, pos, this.gameModeForPlayer)) {
                     this.player.connection.send(new ClientboundBlockUpdatePacket(pos, this.level.getBlockState(pos)));
                     this.debugLogging(pos, false, sequence, "block action restricted");
@@ -181,7 +_,19 @@
                 this.destroyProgressStart = this.gameTicks;
                 float progress = 1.0F;
                 BlockState blockState = this.level.getBlockState(pos);
-                if (!blockState.isAir()) {
+                // CraftBukkit start - Swings at air do *NOT* exist.
+                if (bukkit_event.useInteractedBlock() == Event.Result.DENY) {
+                    // If we denied a door from opening, we need to send a correcting update to the client, as it already opened the door.
+                    BlockState data = this.level.getBlockState(pos);
+                    if (data.getBlock() instanceof DoorBlock) {
+                        // For some reason *BOTH* the bottom/top part have to be marked updated.
+                        boolean bottom = data.getValue(DoorBlock.HALF) == DoubleBlockHalf.LOWER;
+                        this.player.connection.send(new ClientboundBlockUpdatePacket(this.level, pos));
+                        this.player.connection.send(new ClientboundBlockUpdatePacket(this.level, bottom ? pos.above() : pos.below()));
+                    } else if (data.getBlock() instanceof TrapDoorBlock) {
+                        this.player.connection.send(new ClientboundBlockUpdatePacket(this.level, pos));
+                    }
+                } else if (!blockState.isAir()) {
                     EnchantmentHelper.onHitBlock(
                         this.level,
                         this.player.getMainHandItem(),
@@ -192,10 +_,31 @@
                         blockState,
                         item -> this.player.onEquippedItemBroken(item, EquipmentSlot.MAINHAND)
                     );
+                    if (event.getUseBlock() != net.minecraft.util.TriState.FALSE)
                     blockState.attack(this.level, pos, this.player);
                     progress = blockState.getDestroyProgress(this.player, this.player.level(), pos);
                 }
 
+                if (bukkit_event.useItemInHand() == Event.Result.DENY) {
+                    // If we 'insta destroyed' then the client needs to be informed.
+                    if (progress > 1.0f) {
+                        this.player.connection.send(new ClientboundBlockUpdatePacket(this.level, pos));
+                    }
+                    return;
+                }
+                org.bukkit.event.block.BlockDamageEvent blockEvent = CraftEventFactory.callBlockDamageEvent(this.player, pos, this.player.getInventory().getSelectedItem(), progress >= 1.0f);
+
+                if (blockEvent.isCancelled()) {
+                    // Let the client know the block still exists
+                    this.player.connection.send(new ClientboundBlockUpdatePacket(this.level, pos));
+                    return;
+                }
+
+                if (blockEvent.getInstaBreak()) {
+                    progress = 2.0f;
+                }
+                // CraftBukkit end
+
                 if (!blockState.isAir() && progress >= 1.0F) {
                     this.destroyAndAck(pos, sequence, "insta mine");
                 } else {
@@ -237,13 +_,15 @@
             } else if (action == ServerboundPlayerActionPacket.Action.ABORT_DESTROY_BLOCK) {
                 this.isDestroyingBlock = false;
                 if (!Objects.equals(this.destroyPos, pos)) {
-                    LOGGER.warn("Mismatch in destroy block pos: {} {}", this.destroyPos, pos);
+                    LOGGER.debug("Mismatch in destroy block pos: {} {}", this.destroyPos, pos);
                     this.level.destroyBlockProgress(this.player.getId(), this.destroyPos, -1);
                     this.debugLogging(pos, true, sequence, "aborted mismatched destroying");
                 }
 
                 this.level.destroyBlockProgress(this.player.getId(), pos, -1);
                 this.debugLogging(pos, true, sequence, "aborted destroying");
+
+                CraftEventFactory.callBlockDamageAbortEvent(this.player, pos, this.player.getInventory().getSelectedItem()); // CraftBukkit
             }
         }
     }
@@ -258,8 +_,10 @@
     }
 
     public boolean destroyBlock(BlockPos pos) {
+        // Neo: Fire the BlockBreakEvent, and ignore the original ItemStack#canDestroyBlock check since the break event manages the status of it.
         BlockState state = this.level.getBlockState(pos);
-        if (!this.player.getMainHandItem().canDestroyBlock(state, this.level, pos, this.player)) {
+        var event = net.neoforged.neoforge.common.CommonHooks.fireBlockBreak(level, gameModeForPlayer, player, pos, state);
+        if (event.isCanceled()) {
             return false;
         }
 
@@ -275,28 +_,44 @@
         }
 
         BlockState adjustedState = block.playerWillDestroy(this.level, pos, state, this.player);
-        boolean changed = this.level.removeBlock(pos, false);
-        if (SharedConstants.DEBUG_BLOCK_BREAK) {
-            LOGGER.info("server broke {} {} -> {}", pos, adjustedState, this.level.getBlockState(pos));
-        }
-
-        if (changed) {
-            block.destroy(this.level, pos, adjustedState);
-        }
 
         if (this.player.preventsBlockDrops()) {
+            removeBlock(pos, adjustedState, false, player.getMainHandItem().copy());
             return true;
         }
 
         ItemStack itemStack = this.player.getMainHandItem();
         ItemStack destroyedWith = itemStack.copy();
-        boolean canDestroy = this.player.hasCorrectToolForDrops(adjustedState);
+        boolean canDestroy = adjustedState.canHarvestBlock(this.level, pos, this.player); // previously player.hasCorrectToolForDrops(blockstate)
         itemStack.mineBlock(this.level, adjustedState, pos, this.player);
+        boolean changed = removeBlock(pos, adjustedState, canDestroy, destroyedWith.copy());
+
         if (changed && canDestroy) {
             block.playerDestroy(this.level, this.player, pos, adjustedState, blockEntity, destroyedWith);
         }
 
+        // Neo: Fire the PlayerDestroyItemEvent if the tool was broken at any point during the break process
+        if (itemStack.isEmpty() && !destroyedWith.isEmpty()) {
+            net.neoforged.neoforge.event.EventHooks.onPlayerDestroyItem(this.player, destroyedWith, InteractionHand.MAIN_HAND);
+        }
+
         return true;
+    }
+
+    /**
+     * Neo: Patched-in method that handles actual removal of blocks for {@link #destroyBlock(BlockPos)}.
+     *
+     * @param pos The block pos of the destroyed block
+     * @param state The state of the destroyed block
+     * @param canHarvest If the player breaking the block can harvest the drops of the block
+     * @param toolStack The players main-hand prior to destroying the block and applying damage to the tool.
+     * @return If the block was removed, as reported by {@link BlockState#onDestroyedByPlayer}.
+     */
+    private boolean removeBlock(BlockPos pos, BlockState state, boolean canHarvest, ItemStack toolStack) {
+        boolean removed = state.onDestroyedByPlayer(this.level, pos, this.player, toolStack, canHarvest, this.level.getFluidState(pos));
+        if (removed)
+            state.getBlock().destroy(this.level, pos, state);
+        return removed;
     }
 
     public InteractionResult useItem(ServerPlayer player, Level level, ItemStack itemStack, InteractionHand hand) {
@@ -308,6 +_,9 @@
             return InteractionResult.PASS;
         }
 
+        InteractionResult cancelResult = net.neoforged.neoforge.common.CommonHooks.onItemRightClick(player, hand);
+        if (cancelResult != null) return cancelResult;
+
         int oldCount = itemStack.getCount();
         int oldDamage = itemStack.getDamageValue();
         InteractionResult result = itemStack.use(level, player, hand);
@@ -344,6 +_,13 @@
         return result;
     }
 
+
+    // CraftBukkit start - whole method
+    public boolean interactResult = false;
+    public boolean firedInteract = false;
+    public BlockPos interactPosition;
+    public InteractionHand interactHand;
+    public ItemStack interactItemStack;
     public InteractionResult useItemOn(ServerPlayer player, Level level, ItemStack itemStack, InteractionHand hand, BlockHitResult hitResult) {
         BlockPos pos = hitResult.getBlockPos();
         BlockState state = level.getBlockState(pos);
@@ -351,6 +_,8 @@
             return InteractionResult.FAIL;
         }
 
+        net.neoforged.neoforge.event.entity.player.PlayerInteractEvent.RightClickBlock event = net.neoforged.neoforge.common.CommonHooks.onRightClickBlock(player, hand, pos, hitResult);
+        if (event.isCanceled()) return event.getCancellationResult();
         if (this.gameModeForPlayer == GameType.SPECTATOR) {
             MenuProvider menuProvider = state.getMenuProvider(level, pos);
             if (menuProvider != null) {
@@ -360,10 +_,15 @@
                 return InteractionResult.PASS;
             }
         } else {
+            UseOnContext context = new UseOnContext(player, hand, hitResult);
+            if (event.getUseItem() != net.minecraft.util.TriState.FALSE) {
+                InteractionResult result = itemStack.onItemUseFirst(context);
+                if (result != InteractionResult.PASS) return result;
+            }
             boolean haveSomethingInOurHands = !player.getMainHandItem().isEmpty() || !player.getOffhandItem().isEmpty();
-            boolean suppressUsingBlock = player.isSecondaryUseActive() && haveSomethingInOurHands;
+            boolean suppressUsingBlock = (player.isSecondaryUseActive() && haveSomethingInOurHands) && !(player.getMainHandItem().doesSneakBypassUse(level, pos, player) && player.getOffhandItem().doesSneakBypassUse(level, pos, player));
             ItemStack usedItemStack = itemStack.copy();
-            if (!suppressUsingBlock) {
+            if (event.getUseBlock().isTrue() || (event.getUseBlock().isDefault() && !suppressUsingBlock)) {
                 InteractionResult itemUse = state.useItemOn(player.getItemInHand(hand), level, player, hand, hitResult);
                 if (itemUse.consumesAction()) {
                     CriteriaTriggers.ITEM_USED_ON_BLOCK.trigger(player, pos, usedItemStack);
@@ -379,8 +_,8 @@
                 }
             }
 
-            if (!itemStack.isEmpty() && !player.getCooldowns().isOnCooldown(itemStack)) {
-                UseOnContext context = new UseOnContext(player, hand, hitResult);
+            if (event.getUseItem().isTrue() || (!itemStack.isEmpty() && !player.getCooldowns().isOnCooldown(itemStack))) {
+                if (event.getUseItem().isFalse()) return InteractionResult.PASS;
                 InteractionResult success;
                 if (player.hasInfiniteMaterials()) {
                     int count = itemStack.getCount();
