--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -1,6 +_,7 @@
 package net.minecraft.server.level;
 
 import com.google.common.annotations.VisibleForTesting;
+import com.mohistmc.youer.util.IterableWrapper;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
 import it.unimi.dsi.fastutil.longs.LongSet;
@@ -15,9 +_,10 @@
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.BooleanSupplier;
 import java.util.function.Consumer;
-import java.util.function.Predicate;
+import java.util.function.Supplier;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
 import net.minecraft.network.protocol.Packet;
@@ -56,17 +_,19 @@
 
 import net.neoforged.neoforge.common.extensions.IServerChunkCacheExtension;
 
-public class ServerChunkCache extends ChunkSource implements IServerChunkCacheExtension {
+public class ServerChunkCache extends ChunkSource implements IServerChunkCacheExtension, ca.spottedleaf.moonrise.patches.chunk_system.world.ChunkSystemServerChunkCache { // Paper - rewrite chunk system
     private static final Logger LOGGER = LogUtils.getLogger();
     private final DistanceManager distanceManager;
     public final ServerLevel level;
-    private final Thread mainThread;
+    public final Thread mainThread;
     private final ThreadedLevelLightEngine lightEngine;
     public final ServerChunkCache.MainThreadExecutor mainThreadProcessor;
     public final ChunkMap chunkMap;
     private final SavedDataStorage savedDataStorage;
-    private final TicketStorage ticketStorage;
+    public final TicketStorage ticketStorage;
+    private long lastInhabitedUpdate;
     public boolean spawnEnemies = true;
+    public boolean spawnFriendlies = true; // Paper - add back spawnFriendlies field
     private static final int CACHE_SIZE = 4;
     private final long[] lastChunkPos = new long[4];
     private final @Nullable ChunkStatus[] lastChunkStatus = new ChunkStatus[4];
@@ -75,7 +_,115 @@
     private final Set<ChunkHolder> chunkHoldersToBroadcast = new ReferenceOpenHashSet<>();
     @VisibleForDebug
     private NaturalSpawner.@Nullable SpawnState lastSpawnState;
-
+    // Paper start
+    public final ca.spottedleaf.concurrentutil.map.concurrent.longs.ConcurrentChainedLong2ReferenceHashTable<net.minecraft.world.level.chunk.LevelChunk> fullChunks = new ca.spottedleaf.concurrentutil.map.concurrent.longs.ConcurrentChainedLong2ReferenceHashTable<>();
+    public int getFullChunksCount() {
+        return this.fullChunks.size();
+    }
+    long chunkFutureAwaitCounter;
+    // Paper end
+    // Paper start - rewrite chunk system
+
+    @Override
+    public final void moonrise$setFullChunk(final int chunkX, final int chunkZ, final LevelChunk chunk) {
+        final long key = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ);
+        if (chunk == null) {
+            this.fullChunks.remove(key);
+        } else {
+            this.fullChunks.put(key, chunk);
+        }
+    }
+
+    @Override
+    public final LevelChunk moonrise$getFullChunkIfLoaded(final int chunkX, final int chunkZ) {
+        return this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
+    }
+
+    private ChunkAccess syncLoad(final int chunkX, final int chunkZ, final ChunkStatus toStatus) {
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler chunkTaskScheduler = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler();
+        final CompletableFuture<ChunkAccess> completable = new CompletableFuture<>();
+        chunkTaskScheduler.scheduleChunkLoad(
+                chunkX, chunkZ, toStatus, true, ca.spottedleaf.concurrentutil.util.Priority.BLOCKING,
+                completable::complete
+        );
+
+        if (!completable.isDone() && chunkTaskScheduler.hasShutdown()) {
+            throw new IllegalStateException(
+                    "Chunk system has shut down, cannot process chunk requests in world '" + ca.spottedleaf.moonrise.common.util.WorldUtil.getWorldName(this.level) + "' at "
+                            + "(" + chunkX + "," + chunkZ + ") status: " + toStatus
+            );
+        }
+
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, chunkX, chunkZ)) {
+            ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.pushChunkWait(this.level, chunkX, chunkZ);
+            this.mainThreadProcessor.managedBlock(completable::isDone);
+            ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.popChunkWait();
+        }
+
+        final ChunkAccess ret = completable.join();
+        if (ret == null) {
+            throw new IllegalStateException("Chunk not loaded when requested");
+        }
+
+        return ret;
+    }
+
+    private ChunkAccess getChunkFallback(final int chunkX, final int chunkZ, final ChunkStatus toStatus,
+                                         final boolean load) {
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler chunkTaskScheduler = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler();
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkHolderManager chunkHolderManager = chunkTaskScheduler.chunkHolderManager;
+
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder currentChunk = chunkHolderManager.getChunkHolder(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
+
+        final ChunkAccess ifPresent = currentChunk == null ? null : currentChunk.getChunkIfPresent(toStatus);
+
+        if (ifPresent != null && (toStatus != ChunkStatus.FULL || currentChunk.isFullChunkReady())) {
+            return ifPresent;
+        }
+
+        final ca.spottedleaf.moonrise.common.PlatformHooks platformHooks = ca.spottedleaf.moonrise.common.PlatformHooks.get();
+
+        if (platformHooks.hasCurrentlyLoadingChunk() && currentChunk != null) {
+            final ChunkAccess loading = platformHooks.getCurrentlyLoadingChunk(currentChunk.vanillaChunkHolder);
+            if (loading != null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) {
+                return loading;
+            }
+        }
+
+        return load ? this.syncLoad(chunkX, chunkZ, toStatus) : null;
+    }
+    // Paper end - rewrite chunk system
+    // Paper start - chunk tick iteration optimisations
+    private final ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom shuffleRandom = new ca.spottedleaf.moonrise.common.util.SimpleThreadUnsafeRandom(0L);
+    private void iterateTickingChunksFaster() {
+        final ServerLevel world = this.level;
+        final int randomTickSpeed = world.getGameRules().get(GameRules.RANDOM_TICK_SPEED);
+
+        // TODO check on update: impl of forEachBlockTickingChunk will only iterate ENTITY ticking chunks!
+        // TODO check on update: consumer just runs tickChunk
+        final ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.world.level.chunk.LevelChunk> entityTickingChunks = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getEntityTickingChunks();
+
+        // note: we can use the backing array here because:
+        // 1. we do not care about new additions
+        // 2. _removes_ are impossible at this stage in the tick
+        final LevelChunk[] raw = entityTickingChunks.getRawDataUnchecked();
+        final int size = entityTickingChunks.size();
+
+        java.util.Objects.checkFromToIndex(0, size, raw.length);
+        for (int i = 0; i < size; ++i) {
+            world.tickChunk(raw[i], randomTickSpeed);
+
+            // call mid-tick tasks for chunk system
+            if ((i & 7) == 0) {
+                // ((ca.spottedleaf.moonrise.patches.chunk_system.server.ChunkSystemMinecraftServer)this.level.getServer()).moonrise$executeMidTickTasks();
+                continue;
+            }
+        }
+    }
+    // Paper end - chunk tick iteration optimisations
+
+    public boolean pufferfish$firstRunSpawnCounts = true; // Pufferfish
+    public final java.util.concurrent.atomic.AtomicBoolean pufferfish$spawnCountsReady = new java.util.concurrent.atomic.AtomicBoolean(false); // Pufferfish - Optimize count mobs
     public ServerChunkCache(
         ServerLevel level,
         LevelStorageSource.LevelStorageAccess levelStorage,
@@ -86,7 +_,9 @@
         int viewDistance,
         int simulationDistance,
         boolean syncWrites,
-        ChunkStatusUpdateListener chunkStatusListener
+        ChunkStatusUpdateListener chunkStatusListener,
+        Supplier<SavedDataStorage> overworldDataStorage
+        , final SavedDataStorage savedDataStorage // Paper - initialize SavedDataStorage earlier
     ) {
         this.level = level;
         this.mainThreadProcessor = new ServerChunkCache.MainThreadExecutor(level);
@@ -99,7 +_,9 @@
             LOGGER.error("Failed to create dimension data storage directory", e);
         }
 
-        this.savedDataStorage = new SavedDataStorage(dataFolder, fixerUpper, level.registryAccess());
+        this.savedDataStorage = savedDataStorage; // Paper - initialize SavedDataStorage earlier
+        // Youer: Make level context available to SavedDataType factories that depend on it
+        savedDataStorage.setLevel(level);
         this.ticketStorage = this.savedDataStorage.computeIfAbsent(TicketStorage.TYPE);
         this.chunkMap = new ChunkMap(
             level,
@@ -111,6 +_,7 @@
             this,
             generator,
             chunkStatusListener,
+            overworldDataStorage,
             this.ticketStorage,
             viewDistance,
             syncWrites
@@ -121,6 +_,57 @@
         this.clearCache();
     }
 
+    // CraftBukkit start - properly implement isChunkLoaded
+    public boolean isChunkLoaded(int chunkX, int chunkZ) {
+        ChunkHolder chunk = this.chunkMap.getUpdatingChunkIfPresent(ChunkPos.pack(chunkX, chunkZ));
+        if (chunk == null) {
+            return false;
+        }
+        return chunk.getFullChunkNow() != null;
+    }
+    // CraftBukkit end
+    // Paper - rewrite chunk system
+
+    @Nullable
+    public ChunkAccess getChunkAtImmediately(int x, int z) {
+        ChunkHolder holder = this.chunkMap.getVisibleChunkIfPresent(ChunkPos.pack(x, z));
+        if (holder == null) {
+            return null;
+        }
+
+        return holder.getLatestChunk();
+    }
+
+    public void addTicketAtLevel(TicketType ticketType, ChunkPos chunkPos, int ticketLevel) {
+        this.ticketStorage.addTicket(new Ticket(ticketType, ticketLevel), chunkPos);
+    }
+
+    public void removeTicketAtLevel(TicketType ticketType, ChunkPos chunkPos, int ticketLevel) {
+        this.ticketStorage.removeTicket(new Ticket(ticketType, ticketLevel), chunkPos);
+    }
+
+    // "real" get chunk if loaded
+    // Note: Partially copied from the getChunkAt method below
+    @Nullable
+    public LevelChunk getChunkAtIfCachedImmediately(int x, int z) {
+        long key = ChunkPos.pack(x, z);
+
+        // Note: Bypass cache since we need to check ticket level, and to make this MT-Safe
+
+        ChunkHolder playerChunk = this.getVisibleChunkIfPresent(key);
+        if (playerChunk == null) {
+            return null;
+        }
+
+        return playerChunk.getFullChunkNowUnchecked();
+    }
+
+    @Nullable
+    public LevelChunk getChunkAtIfLoadedImmediately(int x, int z) {
+        return this.fullChunks.get(ChunkPos.pack(x, z));
+    }
+    // Paper end
+
     public ThreadedLevelLightEngine getLightEngine() {
         return this.lightEngine;
     }
@@ -143,65 +_,41 @@
 
     @Override
     public @Nullable ChunkAccess getChunk(int x, int z, ChunkStatus targetStatus, boolean loadOrGenerate) {
-        if (Thread.currentThread() != this.mainThread) {
-            return CompletableFuture.<ChunkAccess>supplyAsync(() -> this.getChunk(x, z, targetStatus, loadOrGenerate), this.mainThreadProcessor).join();
-        }
-
-        ProfilerFiller profiler = Profiler.get();
-        profiler.incrementCounter("getChunk");
-        long pos = ChunkPos.pack(x, z);
-
-        for (int i = 0; i < 4; i++) {
-            if (pos == this.lastChunkPos[i] && targetStatus == this.lastChunkStatus[i]) {
-                ChunkAccess chunkAccess = this.lastChunk[i];
-                if (chunkAccess != null || !loadOrGenerate) {
-                    return chunkAccess;
-                }
+        // Paper start - rewrite chunk system
+        if (targetStatus == ChunkStatus.FULL) {
+            final LevelChunk ret = this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z));
+
+            if (ret != null) {
+                return ret;
             }
-        }
-
-        profiler.incrementCounter("getChunkCacheMiss");
-        CompletableFuture<ChunkResult<ChunkAccess>> serverFuture = this.getChunkFutureMainThread(x, z, targetStatus, loadOrGenerate);
-        this.mainThreadProcessor.managedBlock(serverFuture::isDone);
-        ChunkResult<ChunkAccess> chunkResult = serverFuture.join();
-        ChunkAccess chunk = chunkResult.orElse(null);
-        if (chunk == null && loadOrGenerate) {
-            throw (IllegalStateException)Util.pauseInIde(new IllegalStateException("Chunk not there when requested: " + chunkResult.getError()));
-        }
-
-        this.storeInCache(pos, chunk, targetStatus);
-        return chunk;
+
+            return loadOrGenerate ? this.getChunkFallback(x, z, targetStatus, loadOrGenerate) : null;
+        }
+
+        return this.getChunkFallback(x, z, targetStatus, loadOrGenerate);
+        // Paper end - rewrite chunk system
     }
 
     @Override
     public @Nullable LevelChunk getChunkNow(int x, int z) {
-        if (Thread.currentThread() != this.mainThread) {
-            return null;
-        }
-
-        Profiler.get().incrementCounter("getChunkNow");
-        long pos = ChunkPos.pack(x, z);
-
-        for (int i = 0; i < 4; i++) {
-            if (pos == this.lastChunkPos[i] && this.lastChunkStatus[i] == ChunkStatus.FULL) {
-                return this.lastChunk[i] instanceof LevelChunk levelChunk ? levelChunk : null;
-            }
-        }
-
-        ChunkHolder chunkHolder = this.getVisibleChunkIfPresent(pos);
-        if (chunkHolder == null) {
-            return null;
-        }
-
-        ChunkAccess chunk = chunkHolder.getChunkIfPresent(ChunkStatus.FULL);
-        if (chunk != null) {
-            this.storeInCache(pos, chunk, ChunkStatus.FULL);
-            if (chunk instanceof LevelChunk levelChunk) {
-                return levelChunk;
-            }
-        }
-
-        return null;
+        // Paper start - rewrite chunk system
+        final LevelChunk ret = this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z));
+        if (!ca.spottedleaf.moonrise.common.PlatformHooks.get().hasCurrentlyLoadingChunk()) {
+            return ret;
+        }
+
+        if (ret != null || !ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) {
+            return ret;
+        }
+
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder holder = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler()
+                .chunkHolderManager.getChunkHolder(x, z);
+        if (holder == null) {
+            return ret;
+        }
+
+        return ca.spottedleaf.moonrise.common.PlatformHooks.get().getCurrentlyLoadingChunk(holder.vanillaChunkHolder);
+        // Paper end - rewrite chunk system
     }
 
     private void clearCache() {
@@ -227,45 +_,62 @@
     }
 
     private CompletableFuture<ChunkResult<ChunkAccess>> getChunkFutureMainThread(int x, int z, ChunkStatus targetStatus, boolean loadOrGenerate) {
-        ChunkPos pos = new ChunkPos(x, z);
-        long key = pos.pack();
-        int targetTicketLevel = ChunkLevel.byStatus(targetStatus);
-        ChunkHolder chunkHolder = this.getVisibleChunkIfPresent(key);
-        if (loadOrGenerate) {
-            this.addTicket(new Ticket(TicketType.UNKNOWN, targetTicketLevel), pos);
-            if (this.chunkAbsent(chunkHolder, targetTicketLevel)) {
-                ProfilerFiller profiler = Profiler.get();
-                profiler.push("chunkLoad");
-                this.runDistanceManagerUpdates();
-                chunkHolder = this.getVisibleChunkIfPresent(key);
-                profiler.pop();
-                if (this.chunkAbsent(chunkHolder, targetTicketLevel)) {
-                    throw (IllegalStateException)Util.pauseInIde(new IllegalStateException("No chunk holder after ticket has been added"));
+        // Paper start - rewrite chunk system
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.level, x, z, "Scheduling chunk load off-main");
+
+        final int minLevel = ChunkLevel.byStatus(targetStatus);
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler().chunkHolderManager.getChunkHolder(x, z);
+
+        final boolean needsFullScheduling = targetStatus == ChunkStatus.FULL && (chunkHolder == null || !chunkHolder.getChunkStatus().isOrAfter(FullChunkStatus.FULL));
+
+        if ((chunkHolder == null || chunkHolder.getTicketLevel() > minLevel || needsFullScheduling) && !loadOrGenerate) {
+            return ChunkHolder.UNLOADED_CHUNK_FUTURE;
+        }
+
+        final ChunkAccess ifPresent = chunkHolder == null ? null : chunkHolder.getChunkIfPresent(targetStatus);
+        if (needsFullScheduling || ifPresent == null) {
+            // schedule
+            final CompletableFuture<ChunkResult<ChunkAccess>> ret = new CompletableFuture<>();
+            final Consumer<ChunkAccess> complete = (ChunkAccess chunk) -> {
+                if (chunk == null) {
+                    ret.complete(ChunkHolder.UNLOADED_CHUNK);
+                } else {
+                    ret.complete(ChunkResult.of(chunk));
                 }
-            }
+            };
+
+            ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler().scheduleChunkLoad(
+                    x, z, targetStatus, true,
+                    ca.spottedleaf.concurrentutil.util.Priority.HIGHER,
+                    complete
+            );
+
+            return ret;
+        } else {
+            // can return now
+            return CompletableFuture.completedFuture(ChunkResult.of(ifPresent));
         }
-
-        return this.chunkAbsent(chunkHolder, targetTicketLevel)
-            ? GenerationChunkHolder.UNLOADED_CHUNK_FUTURE
-            : chunkHolder.scheduleChunkGenerationTask(targetStatus, this.chunkMap);
+        // Paper end - rewrite chunk system
     }
 
     private boolean chunkAbsent(@Nullable ChunkHolder chunkHolder, int targetTicketLevel) {
-        return chunkHolder == null || chunkHolder.getTicketLevel() > targetTicketLevel;
+        throw new UnsupportedOperationException(); // Paper - rewrite chunk system
     }
 
     @Override
     public boolean hasChunk(int x, int z) {
-        ChunkHolder chunkHolder = this.getVisibleChunkIfPresent(new ChunkPos(x, z).pack());
-        int targetTicketLevel = ChunkLevel.byStatus(ChunkStatus.FULL);
-        return !this.chunkAbsent(chunkHolder, targetTicketLevel);
+        return this.getChunkNow(x, z) != null; // Paper - rewrite chunk system
     }
 
     @Override
     public @Nullable LightChunk getChunkForLighting(int x, int z) {
-        long key = ChunkPos.pack(x, z);
-        ChunkHolder chunkHolder = this.getVisibleChunkIfPresent(key);
-        return chunkHolder == null ? null : chunkHolder.getChunkIfPresentUnchecked(ChunkStatus.INITIALIZE_LIGHT.getParent());
+        // Paper start - rewrite chunk system
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder newChunkHolder = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler().chunkHolderManager.getChunkHolder(x, z);
+        if (newChunkHolder == null) {
+            return null;
+        }
+        return newChunkHolder.getChunkIfPresentUnchecked(ChunkStatus.INITIALIZE_LIGHT.getParent());
+        // Paper end - rewrite chunk system
     }
 
     public Level getLevel() {
@@ -277,50 +_,61 @@
     }
 
     public boolean runDistanceManagerUpdates() {
-        boolean updated = this.distanceManager.runAllUpdates(this.chunkMap);
-        boolean promoted = this.chunkMap.promoteChunkMap();
-        this.chunkMap.runGenerationTasks();
-        if (!updated && !promoted) {
-            return false;
-        }
-
-        this.clearCache();
-        return true;
+        return ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler().chunkHolderManager.processTicketUpdates(); // Paper - rewrite chunk system
     }
 
     public boolean isPositionTicking(long chunkKey) {
-        if (!this.level.shouldTickBlocksAt(chunkKey)) {
-            return false;
-        }
-
-        ChunkHolder holder = this.getVisibleChunkIfPresent(chunkKey);
-        return holder == null ? false : holder.getTickingChunkFuture().getNow(ChunkHolder.UNLOADED_LEVEL_CHUNK).isSuccess();
+        // Paper start - rewrite chunk system
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder newChunkHolder = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler().chunkHolderManager.getChunkHolder(chunkKey);
+        return newChunkHolder != null && newChunkHolder.isTickingReady();
+        // Paper end - rewrite chunk system
     }
 
     public void save(boolean flushStorage) {
-        this.runDistanceManagerUpdates();
+        // Paper - rewrite chunk system
         this.chunkMap.saveAllChunks(flushStorage);
     }
 
     @Override
     public void close() throws IOException {
-        this.save(true);
+        // CraftBukkit start
+        this.close(true);
+    }
+
+    public void close(boolean save) throws IOException {
+        // Paper - rewrite chunk system
+        // CraftBukkit end
         this.savedDataStorage.close();
-        this.lightEngine.close();
-        this.chunkMap.close();
-    }
+        ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler().chunkHolderManager.close(save, true); // Paper - rewrite chunk system
+    }
+
+    // CraftBukkit start - modelled on below
+    public void purgeUnload() {
+        if (true) return; // Paper - rewrite chunk system
+        ProfilerFiller gameprofilerfiller = Profiler.get();
+
+        gameprofilerfiller.push("purge");
+        this.ticketStorage.purgeStaleTickets(this.chunkMap);
+        this.runDistanceManagerUpdates();
+        gameprofilerfiller.popPush("unload");
+        this.chunkMap.tick(() -> true);
+        gameprofilerfiller.pop();
+        this.clearCache();
+    }
+    // CraftBukkit end
 
     @Override
     public void tick(BooleanSupplier haveTime, boolean tickChunks) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("purge");
-        if (this.level.tickRateManager().runsNormally() || !tickChunks) {
+        if (this.level.tickRateManager().runsNormally() || !tickChunks || this.level.spigotConfig.unloadFrozenChunks) { // Spigot
             this.ticketStorage.purgeStaleTickets(this.chunkMap);
         }
 
         this.runDistanceManagerUpdates();
         profiler.popPush("chunks");
         if (tickChunks) {
+            ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getPlayerChunkLoader().tick(); // Paper - rewrite chunk system
             this.tickChunks();
             this.chunkMap.tick();
         }
@@ -332,25 +_,62 @@
     }
 
     private void tickChunks() {
+        long time = this.level.getGameTime();
+        long timeDiff = time - this.lastInhabitedUpdate;
+        this.lastInhabitedUpdate = time;
         if (!this.level.isDebug()) {
             ProfilerFiller profiler = Profiler.get();
             profiler.push("pollingChunks");
-            if (this.level.tickRateManager().runsNormally()) {
+            if (this.level.tickRateManager().runsNormally() && this.level.paperConfig().unsupportedSettings.ticking.chunks) { // Paper - option to disable ticking
                 profiler.push("tickingChunks");
-                this.tickChunks(profiler);
+                this.tickChunks(profiler, timeDiff);
                 profiler.pop();
             }
 
             this.broadcastChangedChunks(profiler);
             profiler.pop();
         }
+        // Pufferfish start - Optimize mob spawning
+        if (true) {
+            for (ServerPlayer player : this.level.players()) {
+                // Paper start - per player mob spawning backoff
+                for (int j = 0; j < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; j++) {
+                    player.mobCounts[j] = 0;
+
+                    int newBackoff = player.mobBackoffCounts[j] - 1; // TODO make configurable bleed // TODO use nonlinear algorithm?
+                    if (newBackoff < 0) {
+                        newBackoff = 0;
+                    }
+                    player.mobBackoffCounts[j] = newBackoff;
+                }
+                // Paper end - per player mob spawning backoff
+            }
+            if (this.pufferfish$firstRunSpawnCounts) {
+                this.pufferfish$firstRunSpawnCounts = false;
+                this.pufferfish$spawnCountsReady.set(true);
+            }
+            if (this.pufferfish$spawnCountsReady.getAndSet(false)) {
+                net.minecraft.server.MinecraftServer.getServer().pufferfish$mobSpawnExecutor.submit(() -> {
+                    int mapped = this.distanceManager.getNaturalSpawnChunkCount();
+                    ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet.Iterator<Entity> objectiterator =
+                            this.level.entityTickList.entities.iterator(ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet.ITERATOR_FLAG_SEE_ADDITIONS);
+                    try {
+                        IterableWrapper<Entity> wrappedIterator = new IterableWrapper<>(objectiterator);
+                        this.lastSpawnState = NaturalSpawner.createState(mapped, wrappedIterator, this::getFullChunk, null, true);
+                    } finally {
+                        objectiterator.finishedIterating();
+                    }
+                    this.pufferfish$spawnCountsReady.set(true);
+                });
+            }
+        } // Pufferfish end - Optimize mob spawning
     }
 
     private void broadcastChangedChunks(ProfilerFiller profiler) {
         profiler.push("broadcast");
 
         for (ChunkHolder chunkHolder : this.chunkHoldersToBroadcast) {
-            LevelChunk chunk = chunkHolder.getTickingChunk();
+            LevelChunk chunk = chunkHolder.getChunkToSend(); // Paper - rewrite chunk system
             if (chunk != null) {
                 chunkHolder.broadcastChanges(chunk);
             }
@@ -360,17 +_,33 @@
         profiler.pop();
     }
 
-    private void tickChunks(ProfilerFiller profiler) {
+    private void tickChunks(ProfilerFiller profiler, long timeDiff) {
         profiler.push("naturalSpawnCount");
         int chunkCount = this.distanceManager.getNaturalSpawnChunkCount();
-        NaturalSpawner.SpawnState spawnCookie = NaturalSpawner.createState(chunkCount, this.level, this::getFullChunk, new LocalMobCapCalculator(this.chunkMap));
-        this.lastSpawnState = spawnCookie;
-        boolean doMobSpawning = this.level.getGameRules().get(GameRules.SPAWN_MOBS);
+        // Paper start - Optional per player mob spawns
+        NaturalSpawner.SpawnState spawnCookie;
+        if ((!this.spawnFriendlies && !this.spawnEnemies) || !this.level.paperConfig().entities.spawning.perPlayerMobSpawns) {
+            // Pufferfish start - async mob spawning
+            this.lastSpawnState = NaturalSpawner.createState(chunkCount, this.level.getAllEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false);
+            this.pufferfish$spawnCountsReady.set(true);
+            // Pufferfish end - async mob spawning
+        }
+        // Paper end - Optional per player mob spawns
+        // this.lastSpawnState = spawnCookie; // Pufferfish - this is managed asynchronously
+        boolean doMobSpawning = this.level.getGameRules().get(GameRules.SPAWN_MOBS) && !this.level.players().isEmpty(); // CraftBukkit
         int tickSpeed = this.level.getGameRules().get(GameRules.RANDOM_TICK_SPEED);
         List<MobCategory> spawningCategories;
-        if (doMobSpawning) {
-            boolean spawnPersistent = this.level.getGameTime() % 400L == 0L;
-            spawningCategories = NaturalSpawner.getFilteredSpawningCategories(spawnCookie, this.spawnEnemies, spawnPersistent);
+        if (doMobSpawning && (this.spawnEnemies || this.spawnFriendlies)) { // Paper
+            // Paper start - PlayerNaturallySpawnCreaturesEvent
+            for (ServerPlayer player : this.level.players()) {
+                int chunkRange = Math.min(level.spigotConfig.mobSpawnRange, player.getBukkitEntity().getViewDistance());
+                chunkRange = Math.min(chunkRange, 8);
+                player.playerNaturallySpawnedEvent = new com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent(player.getBukkitEntity(), (byte) chunkRange);
+                player.playerNaturallySpawnedEvent.callEvent();
+            }
+            // Paper end - PlayerNaturallySpawnCreaturesEvent
+            boolean spawnPersistent = this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) != 0L && this.level.getGameTime() % this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) == 0L; // CraftBukkit
+            spawningCategories = NaturalSpawner.getFilteredSpawningCategories(this.lastSpawnState, this.spawnFriendlies, this.spawnEnemies, spawnPersistent, this.level); // CraftBukkit // todo snapshot - method lost its spawnFriendlies parameter as mojang is moving away further. We can do what we did prior and reintroduce or re-evaluate. // re-added it for now but let's still re-evaluate later // Pufferfish - Async mob spawning
         } else {
             spawningCategories = List.of();
         }
@@ -381,18 +_,21 @@
             profiler.popPush("filteringSpawningChunks");
             this.chunkMap.collectSpawningChunks(spawningChunks);
             profiler.popPush("shuffleSpawningChunks");
-            Util.shuffle(spawningChunks, this.level.getRandom());
+            // Paper start - chunk tick iteration optimisation
+            this.shuffleRandom.setSeed(this.level.getRandom().nextLong());
+            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(spawningChunks, this.level.getRandom()); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
+            // Paper end - chunk tick iteration optimisation
             profiler.popPush("tickSpawningChunks");
 
             for (LevelChunk chunk : spawningChunks) {
-                this.tickSpawningChunk(chunk, spawningCategories, spawnCookie);
+                this.tickSpawningChunk(chunk, timeDiff, spawningCategories, this.lastSpawnState); // Pufferfish - Async mob spawning
             }
         } finally {
             spawningChunks.clear();
         }
 
         profiler.popPush("tickTickingChunks");
-        this.chunkMap.forEachBlockTickingChunk(chunkx -> this.level.tickChunk(chunkx, tickSpeed));
+        this.iterateTickingChunksFaster(); // Paper - chunk tick iteration optimizations
         if (doMobSpawning) {
             profiler.popPush("customSpawners");
             this.level.tickCustomSpawners(this.spawnEnemies);
@@ -401,25 +_,28 @@
         profiler.pop();
     }
 
-    private void tickSpawningChunk(LevelChunk chunk, List<MobCategory> spawningCategories, NaturalSpawner.SpawnState spawnCookie) {
+    private void tickSpawningChunk(LevelChunk chunk, long timeDiff, List<MobCategory> spawningCategories, NaturalSpawner.SpawnState spawnCookie) {
         ChunkPos chunkPos = chunk.getPos();
-        chunk.incrementInhabitedTime();
-        if (this.distanceManager.inEntityTickingRange(chunkPos.pack())) {
+        chunk.incrementInhabitedTime(timeDiff);
+        if (true) { // Paper - rewrite chunk system
             this.level.tickThunder(chunk);
         }
 
         if (!spawningCategories.isEmpty()) {
-            if (this.level.canSpawnEntitiesInChunk(chunkPos)) {
+            if (this.level.getWorldBorder().isWithinBounds(chunkPos) && pufferfish$spawnCountsReady.get() && this.chunkMap.anyPlayerCloseEnoughForSpawning(chunkPos, true)) { // Paper - rewrite chunk system // Pufferfish - Async mob spawning
                 NaturalSpawner.spawnForChunk(this.level, chunk, spawnCookie, spawningCategories);
             }
         }
     }
 
     private void getFullChunk(long chunkKey, Consumer<LevelChunk> output) {
-        ChunkHolder chunkHolder = this.getVisibleChunkIfPresent(chunkKey);
-        if (chunkHolder != null) {
-            chunkHolder.getFullChunkFuture().getNow(ChunkHolder.UNLOADED_LEVEL_CHUNK).ifSuccess(output);
+        // Paper start - rewrite chunk system
+        // note: bypass currentlyLoaded from getChunkNow
+        final LevelChunk fullChunk = this.fullChunks.get(chunkKey);
+        if (fullChunk != null) {
+            output.accept(fullChunk);
         }
+        // Paper end - rewrite chunk system
     }
 
     @Override
@@ -477,19 +_,11 @@
     }
 
     public CompletableFuture<?> addTicketAndLoadWithRadius(TicketType type, ChunkPos pos, int radius) {
-        if (!type.doesLoad()) {
-            throw new IllegalStateException("Ticket type " + type + " does not trigger chunk loading");
-        }
-
-        if (type.canExpireIfUnloaded()) {
-            throw new IllegalStateException("Ticket type " + type + " can expire before it loads, cannot fetch asynchronously");
-        }
-
-        this.addTicketWithRadius(type, pos, radius);
-        this.runDistanceManagerUpdates();
-        ChunkHolder chunkHolder = this.getVisibleChunkIfPresent(pos.pack());
-        Objects.requireNonNull(chunkHolder, "No chunk was scheduled for loading");
-        return this.chunkMap.getChunkRangeFuture(chunkHolder, radius, distance -> ChunkStatus.FULL);
+        // Paper start - rewrite chunk system
+        return ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler().chunkHolderManager.addTicketAndLoadWithRadius(
+                type, pos, radius, ChunkStatus.FULL, ca.spottedleaf.concurrentutil.util.Priority.NORMAL
+        );
+        // Paper end - rewrite chunk system
     }
 
     public void addTicketWithRadius(TicketType type, ChunkPos pos, int radius) {
@@ -539,21 +_,29 @@
         this.chunkMap.sendToTrackingPlayers(entity, packet);
     }
 
-    public void sendToTrackingPlayersFiltered(Entity entity, Packet<? super ClientGamePacketListener> packet, Predicate<ServerPlayer> targetPredicate) {
-        this.chunkMap.sendToTrackingPlayersFiltered(entity, packet, targetPredicate);
-    }
-
     public void setViewDistance(int newDistance) {
         this.chunkMap.setServerViewDistance(newDistance);
     }
 
+    // Paper start - rewrite chunk system
+    public void setSendViewDistance(int viewDistance) {
+        ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getPlayerChunkLoader().setSendDistance(viewDistance);
+    }
+    // Paper end - rewrite chunk system
+
     public void setSimulationDistance(int simulationDistance) {
         this.distanceManager.updateSimulationDistance(simulationDistance);
     }
 
     @Override
     public void setSpawnSettings(boolean spawnEnemies) {
+        // CraftBukkit start
+        this.setSpawnSettings(spawnEnemies, this.spawnFriendlies);
+    }
+    public void setSpawnSettings(boolean spawnEnemies, final boolean spawnFriendlies) {
         this.spawnEnemies = spawnEnemies;
+        this.spawnFriendlies = spawnFriendlies;
+        // CraftBukkit end
     }
 
     public String getChunkDebugData(ChunkPos pos) {
@@ -619,13 +_,13 @@
         }
 
         @Override
-        protected boolean pollTask() {
+        public boolean pollTask() {
             if (ServerChunkCache.this.runDistanceManagerUpdates()) {
                 return true;
             }
 
-            ServerChunkCache.this.lightEngine.tryScheduleUpdate();
-            return super.pollTask();
+            // Paper - rewrite chunk system
+            return super.pollTask() | ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)ServerChunkCache.this.level).moonrise$getChunkTaskScheduler().executeMainThreadTask(); // Paper - rewrite chunk system
         }
     }
 }
