--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -25,7 +_,6 @@
 import net.minecraft.util.Util;
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.attribute.EnvironmentAttributeReader;
-import net.minecraft.world.attribute.EnvironmentAttributeSystem;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.flag.FeatureFlagSet;
@@ -37,7 +_,6 @@
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.EntityBlock;
-import net.minecraft.world.level.block.LevelEvent;
 import net.minecraft.world.level.block.entity.BlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.border.WorldBorder;
@@ -63,8 +_,7 @@
 
 public class WorldGenRegion implements WorldGenLevel {
     private static final Logger LOGGER = LogUtils.getLogger();
-    private static final Identifier WORLDGEN_REGION_RANDOM = Identifier.withDefaultNamespace("worldgen_region_random");
-    private final StaticCache2D<@Nullable ChunkAccess> cache;
+    private final StaticCache2D<GenerationChunkHolder> cache;
     private final ChunkAccess center;
     private final ServerLevel level;
     private final long seed;
@@ -77,18 +_,44 @@
     private final ChunkStep generatingStep;
     private @Nullable Supplier<String> currentlyGenerating;
     private final AtomicLong subTickCount = new AtomicLong();
-    private final EnvironmentAttributeSystem environmentAttributes;
+    private static final Identifier WORLDGEN_REGION_RANDOM = Identifier.withDefaultNamespace("worldgen_region_random");
     private final int centerChunkX;
     private final int centerChunkZ;
     private final int writeRadius;
 
+    // Paper start - rewrite chunk system
+    /**
+     * During feature generation, light data is not initialised and will always return 15 in Starlight. Vanilla
+     * can possibly return 0 if partially initialised, which allows some mushroom blocks to generate.
+     * In general, the brightness value from the light engine should not be used until the chunk is ready. To emulate
+     * Vanilla behavior better, we return 0 as the brightness during world gen unless the target chunk is finished
+     * lighting.
+     */
+    @Override
+    public int getBrightness(final net.minecraft.world.level.LightLayer lightLayer, final BlockPos blockPos) {
+        final ChunkAccess chunk = this.getChunk(blockPos.getX() >> 4, blockPos.getZ() >> 4);
+        if (!chunk.isLightCorrect()) {
+            return 0;
+        }
+        return this.getLightEngine().getLayerListener(lightLayer).getLightValue(blockPos);
+    }
+
+    /**
+     * See above
+     */
+    @Override
+    public int getRawBrightness(final BlockPos blockPos, final int subtract) {
+        final ChunkAccess chunk = this.getChunk(blockPos.getX() >> 4, blockPos.getZ() >> 4);
+        if (!chunk.isLightCorrect()) {
+            return 0;
+        }
+        return this.getLightEngine().getRawBrightness(blockPos, subtract);
+    }
+    // Paper end - rewrite chunk system
+
     public WorldGenRegion(ServerLevel level, StaticCache2D<GenerationChunkHolder> cache, ChunkStep generatingStep, ChunkAccess center) {
         this.generatingStep = generatingStep;
-        this.cache = cache.map((holder, x, z) -> {
-            int distance = center.getPos().getChessboardDistance(x, z);
-            ChunkStatus maxAllowedStatus = distance >= generatingStep.directDependencies().size() ? null : generatingStep.directDependencies().get(distance);
-            return maxAllowedStatus == null ? null : holder.getChunkIfPresentUnchecked(maxAllowedStatus);
-        });
+        this.cache = cache;
         this.center = center;
         this.level = level;
         this.seed = level.getSeed();
@@ -100,7 +_,6 @@
         this.centerChunkX = centerPos.x();
         this.centerChunkZ = centerPos.z();
         this.writeRadius = generatingStep.blockStateWriteRadius();
-        this.environmentAttributes = EnvironmentAttributeSystem.builder().addStaticLayers(this).build();
     }
 
     public boolean isOldChunkAround(ChunkPos pos, int range) {
@@ -127,14 +_,17 @@
         ChunkStatus maxAllowedStatus = distance >= this.generatingStep.directDependencies().size()
             ? null
             : this.generatingStep.directDependencies().get(distance);
-        ChunkAccess chunk;
+        GenerationChunkHolder chunkHolder;
         if (maxAllowedStatus != null) {
-            chunk = this.cache.get(chunkX, chunkZ);
-            if (chunk != null && targetStatus.isOrBefore(maxAllowedStatus)) {
-                return chunk;
+            chunkHolder = this.cache.get(chunkX, chunkZ);
+            if (targetStatus.isOrBefore(maxAllowedStatus)) {
+                ChunkAccess chunk = chunkHolder.getChunkIfPresentUnchecked(maxAllowedStatus);
+                if (chunk != null) {
+                    return chunk;
+                }
             }
         } else {
-            chunk = null;
+            chunkHolder = null;
         }
 
         CrashReport report = CrashReport.forThrowable(
@@ -144,7 +_,7 @@
         category.setDetail("Requested chunk", String.format(Locale.ROOT, "%d, %d", chunkX, chunkZ));
         category.setDetail("Generating status", () -> this.generatingStep.targetStatus().getName());
         category.setDetail("Requested status", targetStatus::getName);
-        category.setDetail("Actual status", () -> chunk == null ? "[out of cache bounds]" : chunk.getPersistedStatus().getName());
+        category.setDetail("Actual status", () -> chunkHolder == null ? "[out of cache bounds]" : chunkHolder.getPersistedStatus().getName());
         category.setDetail("Maximum allowed status", () -> maxAllowedStatus == null ? "null" : maxAllowedStatus.getName());
         category.setDetail("Dependencies", this.generatingStep.directDependencies()::toString);
         category.setDetail("Requested distance", distance);
@@ -158,6 +_,28 @@
         return distance < this.generatingStep.directDependencies().size();
     }
 
+    // Paper start - if loaded util
+    @Nullable
+    @Override
+    public ChunkAccess getChunkIfLoadedImmediately(int x, int z) {
+        return this.getChunk(x, z, ChunkStatus.FULL, false);
+    }
+
+    @Nullable
+    @Override
+    public final BlockState getBlockStateIfLoaded(BlockPos pos) {
+        ChunkAccess chunk = this.getChunkIfLoadedImmediately(pos.getX() >> 4, pos.getZ() >> 4);
+        return chunk == null ? null : chunk.getBlockState(pos);
+    }
+
+    @Nullable
+    @Override
+    public final FluidState getFluidIfLoaded(BlockPos pos) {
+        ChunkAccess chunk = this.getChunkIfLoadedImmediately(pos.getX() >> 4, pos.getZ() >> 4);
+        return chunk == null ? null : chunk.getFluidState(pos);
+    }
+    // Paper end
+
     @Override
     public BlockState getBlockState(BlockPos pos) {
         int chunkX = SectionPos.blockToSectionCoord(pos.getX());
@@ -273,22 +_,32 @@
         return Math.abs(this.centerChunkX - chunkX) <= this.writeRadius && Math.abs(this.centerChunkZ - chunkZ) <= this.writeRadius;
     }
 
+    private boolean hasSetFarWarned = false; // Paper - Buffer OOB setBlock calls
     @Override
     public boolean ensureCanWrite(BlockPos pos) {
         if (!this.isWithinWriteZone(pos)) {
             int chunkX = SectionPos.blockToSectionCoord(pos.getX());
             int chunkZ = SectionPos.blockToSectionCoord(pos.getZ());
-            Util.logAndPauseIfInIde(
-                "Detected setBlock in a far chunk ["
-                    + chunkX
-                    + ", "
-                    + chunkZ
-                    + "], pos: "
-                    + pos
-                    + ", status: "
-                    + this.generatingStep.targetStatus()
-                    + (this.currentlyGenerating == null ? "" : ", currently generating: " + this.currentlyGenerating.get())
-            );
+            // Paper start - Buffer OOB setBlock calls
+            if (!org.purpurmc.purpur.PurpurConfig.loggerSuppressSetBlockFarChunk) // Purpur - Logger settings (suppressing pointless logs)
+            if (!hasSetFarWarned) {
+                Util.logAndPauseIfInIde(
+                        "Detected setBlock in a far chunk ["
+                                + chunkX
+                                + ", "
+                                + chunkZ
+                                + "], pos: "
+                                + pos
+                                + ", status: "
+                                + this.generatingStep.targetStatus()
+                                + (this.currentlyGenerating == null ? "" : ", currently generating: " + this.currentlyGenerating.get())
+                );
+                hasSetFarWarned = true;
+                if (this.getServer() != null && this.getServer().isDebugging()) {
+                    io.papermc.paper.util.TraceUtil.dumpTraceForThread("far setBlock call");
+                }
+            }
+            // Paper end - Buffer OOB setBlock calls
             return false;
         } else if (this.center.isUpgrading()) {
             LevelHeightAccessor levelHeightAccessor = this.center.getHeightAccessorForGeneration();
@@ -306,7 +_,7 @@
 
         ChunkAccess chunk = this.getChunk(pos);
         BlockState oldState = chunk.setBlockState(pos, blockState, updateFlags);
-        if (oldState != null) {
+        if (oldState != null && (updateFlags & Block.UPDATE_SKIP_POI) == 0) { // Paper - temporary flag
             this.level.updatePOIOnBlockStateChange(pos, oldState, blockState);
         }
 
@@ -319,6 +_,17 @@
                     chunk.removeBlockEntity(pos);
                 }
             } else {
+                // Paper start - Clear block entity before setting up a DUMMY block entity
+                // The concept of removing a block entity when the block itself changes is generally lifted
+                // from LevelChunk#setBlockState.
+                // It is however to note that this may only run if the block actually changes.
+                // Otherwise a chest block entity generated by a structure template that is later "updated" to
+                // be waterlogged would remove its existing block entity (see PaperMC/Paper#10750)
+                // This logic is *also* found in LevelChunk#setBlockState.
+                if (oldState != null && !java.util.Objects.equals(oldState.getBlock(), blockState.getBlock())) {
+                    chunk.removeBlockEntity(pos);
+                }
+                // Paper end - Clear block entity before setting up a DUMMY block entity
                 CompoundTag tag = new CompoundTag();
                 tag.putInt("x", pos.getX());
                 tag.putInt("y", pos.getY());
@@ -346,6 +_,14 @@
 
     @Override
     public boolean addFreshEntity(Entity entity) {
+        // CraftBukkit start
+        return this.addFreshEntity(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.DEFAULT);
+    }
+
+    @Override
+    public boolean addFreshEntity(final Entity entity, final org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        // CraftBukkit end
+        if (entity instanceof net.minecraft.world.entity.Mob mob && mob.isSpawnCancelled()) return false;
         int xc = SectionPos.blockToSectionCoord(entity.getBlockX());
         int zc = SectionPos.blockToSectionCoord(entity.getBlockZ());
         this.getChunk(xc, zc).addEntity(entity);
@@ -354,7 +_,7 @@
 
     @Override
     public boolean removeBlock(BlockPos pos, boolean movedByPiston) {
-        return this.setBlockAndUpdate(pos, Blocks.AIR.defaultBlockState());
+        return this.setBlock(pos, Blocks.AIR.defaultBlockState(), 3);
     }
 
     @Override
@@ -449,7 +_,7 @@
     }
 
     @Override
-    public void levelEvent(@Nullable Entity source, @LevelEvent.Value int type, BlockPos pos, int data) {
+    public void levelEvent(@Nullable Entity source, int type, BlockPos pos, int data) {
     }
 
     @Override
@@ -503,6 +_,6 @@
 
     @Override
     public EnvironmentAttributeReader environmentAttributes() {
-        return this.environmentAttributes;
+        return EnvironmentAttributeReader.EMPTY;
     }
 }
