--- a/net/minecraft/network/chat/numbers/StyledFormat.java
+++ b/net/minecraft/network/chat/numbers/StyledFormat.java
@@ -28,7 +_,7 @@
     public static final StyledFormat NO_STYLE = new StyledFormat(Style.EMPTY);
     public static final StyledFormat SIDEBAR_DEFAULT = new StyledFormat(Style.EMPTY.withColor(ChatFormatting.RED));
     public static final StyledFormat PLAYER_LIST_DEFAULT = new StyledFormat(Style.EMPTY.withColor(ChatFormatting.YELLOW));
-    final Style style;
+    public final Style style;
 
     public StyledFormat(Style p_313874_) {
         this.style = p_313874_;
