--- a/net/minecraft/world/item/equipment/Equippable.java
+++ b/net/minecraft/world/item/equipment/Equippable.java
@@ -1,14 +_,15 @@
 package net.minecraft.world.item.equipment;
 
+import com.mohistmc.youer.feature.ban.bans.BanItem;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import java.util.Optional;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.HolderSet;
+import net.minecraft.core.RegistryCodecs;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.network.RegistryFriendlyByteBuf;
 import net.minecraft.network.codec.ByteBufCodecs;
 import net.minecraft.network.codec.StreamCodec;
@@ -18,7 +_,6 @@
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.stats.Stats;
 import net.minecraft.tags.EntityTypeTags;
-import net.minecraft.util.Prediction;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.EntityTypes;
@@ -50,7 +_,7 @@
                 SoundEvent.CODEC.optionalFieldOf("equip_sound", SoundEvents.ARMOR_EQUIP_GENERIC).forGetter(Equippable::equipSound),
                 ResourceKey.codec(EquipmentAssets.ROOT_ID).optionalFieldOf("asset_id").forGetter(Equippable::assetId),
                 Identifier.CODEC.optionalFieldOf("camera_overlay").forGetter(Equippable::cameraOverlay),
-                RegistryCodecs.holderSet(Registries.ENTITY_TYPE).optionalFieldOf("allowed_entities").forGetter(Equippable::allowedEntities),
+                RegistryCodecs.homogeneousList(Registries.ENTITY_TYPE).optionalFieldOf("allowed_entities").forGetter(Equippable::allowedEntities),
                 Codec.BOOL.optionalFieldOf("dispensable", true).forGetter(Equippable::dispensable),
                 Codec.BOOL.optionalFieldOf("swappable", true).forGetter(Equippable::swappable),
                 Codec.BOOL.optionalFieldOf("damage_on_hurt", true).forGetter(Equippable::damageOnHurt),
@@ -127,10 +_,24 @@
     }
 
     public InteractionResult swapWithEquipmentSlot(ItemStack inHand, Player player) {
+        if (BanItem.check(player, inHand)) return InteractionResult.FAIL;
         if (player.canUseSlot(this.slot) && this.canBeEquippedBy(player.typeHolder())) {
             ItemStack inEquipmentSlot = player.getItemBySlot(this.slot);
             if ((!EnchantmentHelper.has(inEquipmentSlot, EnchantmentEffectComponents.PREVENT_ARMOR_CHANGE) || player.isCreative())
                 && !ItemStack.isSameItemSameComponents(inHand, inEquipmentSlot)) {
+                // Paper start
+                final io.papermc.paper.event.player.PlayerSwapWithEquipmentSlotEvent event = new io.papermc.paper.event.player.PlayerSwapWithEquipmentSlotEvent(
+                        (org.bukkit.entity.Player) player.getBukkitEntity(),
+                        org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(inHand),
+                        org.bukkit.craftbukkit.CraftEquipmentSlot.getSlot(this.slot),
+                        org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(inEquipmentSlot)
+                );
+                event.callEvent();
+                if (event.isCancelled()) {
+                    ((org.bukkit.entity.Player) player.getBukkitEntity()).updateInventory();
+                    return InteractionResult.FAIL;
+                }
+                // Paper end
                 if (!player.level().isClientSide()) {
                     player.awardStat(Stats.ITEM_USED.get(inHand.getItem()));
                 }
@@ -146,7 +_,7 @@
                 ItemStack swappedToEquipment = inHand.consumeAndReturn(1, player);
                 player.setItemSlot(this.slot, swappedToEquipment);
                 if (!player.getInventory().add(swappedToInventory)) {
-                    player.drop(swappedToInventory, false, Prediction.PREDICTED);
+                    player.drop(swappedToInventory, false);
                 }
 
                 return InteractionResult.SUCCESS.heldItemTransformedTo(inHand);
