--- a/net/minecraft/world/level/blockscan/BoxBlockMatcher.java
+++ b/net/minecraft/world/level/blockscan/BoxBlockMatcher.java
@@ -6,10 +_,17 @@
 import net.minecraft.util.Continuation;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.phys.AABB;
 
 public class BoxBlockMatcher extends BlockMatcher {
     private final BlockPos from;
     private final BlockPos to;
+
+    public BoxBlockMatcher(LevelReader level, AABB box) {
+        BlockPos minPos = BlockPos.containing(box.getMinPosition());
+        BlockPos maxPos = BlockPos.containing(box.getMaxPosition());
+        this(level, minPos, maxPos);
+    }
 
     public BoxBlockMatcher(LevelReader level, BlockPos from, BlockPos to) {
         super(level);
