--- a/net/minecraft/world/level/storage/DerivedLevelData.java
+++ b/net/minecraft/world/level/storage/DerivedLevelData.java
@@ -7,7 +_,7 @@
 
 public class DerivedLevelData implements ServerLevelData {
     private final WorldData worldData;
-    private final ServerLevelData wrapped;
+    public final ServerLevelData wrapped;
 
     public DerivedLevelData(WorldData worldData, ServerLevelData wrapped) {
         this.worldData = worldData;
@@ -80,5 +_,27 @@
     public void fillCrashReportCategory(CrashReportCategory category, LevelHeightAccessor levelHeightAccessor) {
         category.setDetail("Derived", true);
         this.wrapped.fillCrashReportCategory(category, levelHeightAccessor);
+    }
+
+    // Neo: Variable day time code
+
+    @Override
+    public float getDayTimeFraction() {
+        return wrapped.getDayTimeFraction();
+    }
+
+    @Override
+    public float getDayTimePerTick() {
+        return wrapped.getDayTimePerTick();
+    }
+
+    @Override
+    public void setDayTimeFraction(float dayTimeFraction) {
+        wrapped.setDayTimeFraction(dayTimeFraction);
+    }
+
+    @Override
+    public void setDayTimePerTick(float dayTimePerTick) {
+        wrapped.setDayTimePerTick(dayTimePerTick);
     }
 }
