--- a/net/minecraft/world/level/storage/DerivedLevelData.java
+++ b/net/minecraft/world/level/storage/DerivedLevelData.java
@@ -7,7 +_,7 @@
 
 public class DerivedLevelData implements ServerLevelData {
     private final WorldData worldData;
-    private final ServerLevelData wrapped;
+    public final ServerLevelData wrapped;
 
     public DerivedLevelData(WorldData worldData, ServerLevelData wrapped) {
         this.worldData = worldData;
