--- a/net/minecraft/world/level/storage/DerivedLevelData.java
+++ b/net/minecraft/world/level/storage/DerivedLevelData.java
@@ -13,7 +_,7 @@
 
 public class DerivedLevelData implements ServerLevelData {
     private final WorldData worldData;
-    private final ServerLevelData wrapped;
+    public final ServerLevelData wrapped;
 
     public DerivedLevelData(WorldData p_78079_, ServerLevelData p_78080_) {
         this.worldData = p_78079_;
@@ -190,5 +_,27 @@
     public void fillCrashReportCategory(CrashReportCategory p_164852_, LevelHeightAccessor p_164853_) {
         p_164852_.setDetail("Derived", true);
         this.wrapped.fillCrashReportCategory(p_164852_, p_164853_);
+    }
+
+    // Neo: Variable day time code
+
+    @Override
+    public float getDayTimeFraction() {
+        return wrapped.getDayTimeFraction();
+    }
+
+    @Override
+    public float getDayTimePerTick() {
+        return wrapped.getDayTimePerTick();
+    }
+
+    @Override
+    public void setDayTimeFraction(float dayTimeFraction) {
+        wrapped.setDayTimeFraction(dayTimeFraction);
+    }
+
+    @Override
+    public void setDayTimePerTick(float dayTimePerTick) {
+        wrapped.setDayTimePerTick(dayTimePerTick);
     }
 }
