--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -95,7 +_,7 @@
     public static final String ALLOWED_SYMLINKS_CONFIG_NAME = "allowed_symlinks.txt";
     private static final int DISK_SPACE_WARNING_THRESHOLD = 67108864;
     private static final Component LOAD_FOLDER_ACCESS_MESSAGE = Component.translatable("selectWorld.load_folder_access");
-    public final Path baseDir;
+    private final Path baseDir;
     private final Path backupDir;
     private final DataFixer fixerUpper;
     private final DirectoryValidator worldDirValidator;
@@ -145,13 +_,14 @@
         WorldDataConfiguration dataConfiguration,
         Registry<LevelStem> datapackDimensions,
         HolderLookup.Provider registryAccess
+        , final ResourceKey<Level> dimension // Paper - pass dimension
     ) {
         if (DataFixers.getFileFixer().requiresFileFixing(NbtUtils.getDataVersion(levelDataTag))) {
             throw new IllegalStateException("Cannot get level data without file fixing first");
         }
 
         Dynamic<?> dataTag = RegistryOps.injectRegistryContext(levelDataTag, registryAccess);
-        WorldGenSettings worldGenSettings = readExistingSavedData(worldAccess, registryAccess, WorldGenSettings.TYPE)
+        WorldGenSettings worldGenSettings = readExistingSavedData(worldAccess, dimension, registryAccess, WorldGenSettings.TYPE)
             .mapOrElse(
                 Function.identity(),
                 error -> {
@@ -170,9 +_,9 @@
     }
 
     public static <T extends SavedData> DataResult<T> readExistingSavedData(
-        LevelStorageSource.LevelStorageAccess access, HolderLookup.Provider registryAccess, SavedDataType<T> savedDataType
+        LevelStorageSource.LevelStorageAccess access, final ResourceKey<Level> dimension, HolderLookup.Provider registryAccess, SavedDataType<T> savedDataType
     ) {
-        Path dataLocation = savedDataType.id().withSuffix(".dat").resolveAgainst(access.getLevelPath(LevelResource.DATA));
+        Path dataLocation = savedDataType.id().withSuffix(".dat").resolveAgainst(access.getDimensionPath(dimension).resolve(LevelResource.DATA.id())); // Paper - use dimension scoped data (we 'demote' some data from global and add our own dimension scoped data)
 
         CompoundTag fileContents;
         try {
@@ -395,13 +_,28 @@
         if (!validationResults.isEmpty()) {
             throw new ContentValidationException(levelPath, validationResults);
         } else {
-            return new LevelStorageSource.LevelStorageAccess(levelId, levelPath);
-        }
-    }
-
-    public LevelStorageSource.LevelStorageAccess createAccess(String levelId) throws IOException {
-        Path levelPath = this.getLevelPath(levelId);
-        return new LevelStorageSource.LevelStorageAccess(levelId, levelPath);
+            return new LevelStorageSource.LevelStorageAccess(levelId, levelPath, LevelStem.OVERWORLD); // CraftBukkit
+        }
+    }
+
+    public LevelStorageSource.LevelStorageAccess validateAndCreateAccess(String levelId, ResourceKey<LevelStem> dimensionType) throws IOException, ContentValidationException { // CraftBukkit
+        Path levelPath = this.getLevelPath(levelId);
+        List<ForbiddenSymlinkInfo> validationResults = this.worldDirValidator.validateDirectory(levelPath, true);
+        if (!validationResults.isEmpty()) {
+            throw new ContentValidationException(levelPath, validationResults);
+        } else {
+            return new LevelStorageSource.LevelStorageAccess(levelId, levelPath, dimensionType); // CraftBukkit
+        }
+    }
+
+    public LevelStorageSource.LevelStorageAccess createAccess(String levelId) throws IOException { // CraftBukkit
+        Path levelPath = this.getLevelPath(levelId);
+        return new LevelStorageSource.LevelStorageAccess(levelId, levelPath, LevelStem.OVERWORLD); // CraftBukkit
+    }
+
+    public LevelStorageSource.LevelStorageAccess createAccess(String levelId, ResourceKey<LevelStem> dimensionType) throws IOException { // CraftBukkit
+        Path levelPath = this.getLevelPath(levelId);
+        return new LevelStorageSource.LevelStorageAccess(levelId, levelPath, dimensionType); // CraftBukkit
     }
 
     public DirectoryValidator getWorldDirValidator() {
@@ -453,13 +_,31 @@
         }
     }
 
+    // CraftBukkit start
+    public static Path getPre261StorageFolder(Path path, ResourceKey<LevelStem> dimensionType) {
+        if (dimensionType == LevelStem.OVERWORLD) {
+            return path;
+        } else if (dimensionType == LevelStem.NETHER) {
+            return path.resolve("DIM-1");
+        } else if (dimensionType == LevelStem.END) {
+            return path.resolve("DIM1");
+        } else {
+            return path.resolve("dimensions").resolve(dimensionType.identifier().getNamespace()).resolve(dimensionType.identifier().getPath());
+        }
+    }
+    // CraftBukkit end
+
     public class LevelStorageAccess implements AutoCloseable {
         private DirectoryLock lock;
         public final LevelStorageSource.LevelDirectory levelDirectory;
         private final String levelId;
         private final Map<LevelResource, Path> resources = Maps.newHashMap();
+        // CraftBukkit start
+        public final ResourceKey<LevelStem> dimensionType;
 
-        private LevelStorageAccess(String levelId, Path path) throws IOException {
+        private LevelStorageAccess(String levelId, Path path, ResourceKey<LevelStem> dimensionType) throws IOException {
+            this.dimensionType = dimensionType;
+            // CraftBukkit end
             this.levelId = levelId;
             this.levelDirectory = new LevelStorageSource.LevelDirectory(path);
             this.createLock();
@@ -516,7 +_,8 @@
         }
 
         public Path getDimensionPath(ResourceKey<Level> name) {
-            return DimensionType.getStorageFolder(name, this.levelDirectory.path());
+            if (this.dimensionType != null) name = net.minecraft.core.registries.Registries.levelStemToLevel(this.dimensionType);
+            return DimensionType.getStorageFolder(name, this.levelDirectory.path()); // CraftBukkit
         }
 
         private void checkLock() {
@@ -525,6 +_,19 @@
             }
         }
 
+        // Neo: For reading NeoForge's additional save data (mod list, etc.)
+        public void readAdditionalLevelSaveData(boolean fallback) {
+            checkLock();
+            Path path = fallback ? this.levelDirectory.oldDataFile() : this.levelDirectory.dataFile();
+            try {
+                var tag = readLightweightData(path);
+                if (tag instanceof CompoundTag compoundTag)
+                    net.neoforged.neoforge.common.CommonHooks.readAdditionalLevelSaveData(compoundTag, this.levelDirectory);
+            } catch (Exception e) {
+                LOGGER.error("Exception reading {}", path, e);
+            }
+        }
+
         public PlayerDataStorage createPlayerStorage() {
             this.checkLock();
             return new PlayerDataStorage(this, LevelStorageSource.this.fixerUpper);
@@ -554,10 +_,12 @@
             Dynamic<?> unfixedDataTag;
             try {
                 unfixedDataTag = this.getUnfixedDataTag(false);
+                this.readAdditionalLevelSaveData(false);
             } catch (IOException | NbtException | ReportedNbtException e) {
                 LevelStorageSource.LOGGER.warn("Failed to load world data from {}", this.levelDirectory.dataFile(), e);
                 LevelStorageSource.LOGGER.info("Attempting to use fallback {}", this.levelDirectory.oldDataFile());
                 unfixedDataTag = this.getUnfixedDataTag(true);
+                this.readAdditionalLevelSaveData(true);
                 this.restoreLevelDataFromOld();
             }
 
@@ -583,6 +_,7 @@
             CompoundTag dataTag = levelData.createTag(singleplayerUUID);
             CompoundTag root = new CompoundTag();
             root.put("Data", dataTag);
+            net.neoforged.neoforge.common.CommonHooks.writeAdditionalLevelSaveData(levelData, root);
             this.saveLevelData(root);
         }
 
@@ -609,6 +_,10 @@
 
         public Optional<Path> getIconFile() {
             return !this.lock.isValid() ? Optional.empty() : Optional.of(this.levelDirectory.iconFile());
+        }
+
+        public Path getWorldDir() {
+            return baseDir;
         }
 
         public void deleteLevel() throws IOException {
