--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -392,19 +_,19 @@
         return this.backupDir;
     }
 
-    public LevelStorageSource.LevelStorageAccess validateAndCreateAccess(String levelId) throws IOException, ContentValidationException {
+    public LevelStorageSource.LevelStorageAccess validateAndCreateAccess(String levelId, ResourceKey<LevelStem> dimensionType) throws IOException, ContentValidationException {
         Path levelPath = this.getLevelPath(levelId);
         List<ForbiddenSymlinkInfo> validationResults = this.worldDirValidator.validateDirectory(levelPath, true);
         if (!validationResults.isEmpty()) {
             throw new ContentValidationException(levelPath, validationResults);
         } else {
-            return new LevelStorageSource.LevelStorageAccess(levelId, levelPath);
+            return new LevelStorageSource.LevelStorageAccess(levelId, levelPath, dimensionType);
         }
     }
 
-    public LevelStorageSource.LevelStorageAccess createAccess(String levelId) throws IOException {
+    public LevelStorageSource.LevelStorageAccess createAccess(String levelId, ResourceKey<LevelStem> dimensionType) throws IOException {
         Path levelPath = this.getLevelPath(levelId);
-        return new LevelStorageSource.LevelStorageAccess(levelId, levelPath);
+        return new LevelStorageSource.LevelStorageAccess(levelId, levelPath, dimensionType);
     }
 
     public DirectoryValidator getWorldDirValidator() {
@@ -456,13 +_,31 @@
         }
     }
 
+    // CraftBukkit start
+    public static Path getPre261StorageFolder(Path path, ResourceKey<LevelStem> dimensionType) {
+        if (dimensionType == LevelStem.OVERWORLD) {
+            return path;
+        } else if (dimensionType == LevelStem.NETHER) {
+            return path.resolve("DIM-1");
+        } else if (dimensionType == LevelStem.END) {
+            return path.resolve("DIM1");
+        } else {
+            return path.resolve("dimensions").resolve(dimensionType.identifier().getNamespace()).resolve(dimensionType.identifier().getPath());
+        }
+    }
+    // CraftBukkit end
+
     public class LevelStorageAccess implements AutoCloseable {
         private DirectoryLock lock;
         public final LevelStorageSource.LevelDirectory levelDirectory;
         private final String levelId;
         private final Map<LevelResource, Path> resources;
+        // CraftBukkit start
+        public final ResourceKey<LevelStem> dimensionType;
 
-        private LevelStorageAccess(String levelId, Path path) throws IOException {
+        private LevelStorageAccess(String levelId, Path path, ResourceKey<LevelStem> dimensionType) throws IOException {
+            this.dimensionType = dimensionType;
+            // CraftBukkit end
             Objects.requireNonNull(LevelStorageSource.this);
             super();
             this.resources = Maps.newHashMap();
@@ -531,6 +_,19 @@
             }
         }
 
+        // Neo: For reading NeoForge's additional save data (mod list, etc.)
+        public void readAdditionalLevelSaveData(boolean fallback) {
+            checkLock();
+            Path path = fallback ? this.levelDirectory.oldDataFile() : this.levelDirectory.dataFile();
+            try {
+                var tag = readLightweightData(path);
+                if (tag instanceof CompoundTag compoundTag)
+                    net.neoforged.neoforge.common.CommonHooks.readAdditionalLevelSaveData(compoundTag, this.levelDirectory);
+            } catch (Exception e) {
+                LOGGER.error("Exception reading {}", path, e);
+            }
+        }
+
         public PlayerDataStorage createPlayerStorage() {
             this.checkLock();
             return new PlayerDataStorage(this, LevelStorageSource.this.fixerUpper);
@@ -560,10 +_,12 @@
             Dynamic<?> unfixedDataTag;
             try {
                 unfixedDataTag = this.getUnfixedDataTag(false);
+                this.readAdditionalLevelSaveData(false);
             } catch (NbtException | ReportedNbtException | IOException var3) {
                 LevelStorageSource.LOGGER.warn("Failed to load world data from {}", this.levelDirectory.dataFile(), var3);
                 LevelStorageSource.LOGGER.info("Attempting to use fallback {}", this.levelDirectory.oldDataFile());
                 unfixedDataTag = this.getUnfixedDataTag(true);
+                this.readAdditionalLevelSaveData(true);
                 this.restoreLevelDataFromOld();
             }
 
@@ -589,6 +_,7 @@
             CompoundTag dataTag = levelData.createTag(singleplayerUUID);
             CompoundTag root = new CompoundTag();
             root.put("Data", dataTag);
+            net.neoforged.neoforge.common.CommonHooks.writeAdditionalLevelSaveData(levelData, root);
             this.saveLevelData(root);
         }
 
@@ -615,6 +_,10 @@
 
         public Optional<Path> getIconFile() {
             return !this.lock.isValid() ? Optional.empty() : Optional.of(this.levelDirectory.iconFile());
+        }
+
+        public Path getWorldDir() {
+            return baseDir;
         }
 
         public void deleteLevel() throws IOException {
