--- a/net/minecraft/world/level/storage/LevelStorageSource.java
+++ b/net/minecraft/world/level/storage/LevelStorageSource.java
@@ -95,7 +_,7 @@
     public static final String ALLOWED_SYMLINKS_CONFIG_NAME = "allowed_symlinks.txt";
     private static final int DISK_SPACE_WARNING_THRESHOLD = 67108864;
     private static final Component LOAD_FOLDER_ACCESS_MESSAGE = Component.translatable("selectWorld.load_folder_access");
-    public final Path baseDir;
+    private final Path baseDir;
     private final Path backupDir;
     private final DataFixer fixerUpper;
     private final DirectoryValidator worldDirValidator;
@@ -145,13 +_,14 @@
         WorldDataConfiguration dataConfiguration,
         Registry<LevelStem> datapackDimensions,
         HolderLookup.Provider registryAccess
+        , final ResourceKey<Level> dimension // Paper - pass dimension
     ) {
         if (DataFixers.getFileFixer().requiresFileFixing(NbtUtils.getDataVersion(levelDataTag))) {
             throw new IllegalStateException("Cannot get level data without file fixing first");
         }
 
         Dynamic<?> dataTag = RegistryOps.injectRegistryContext(levelDataTag, registryAccess);
-        WorldGenSettings worldGenSettings = readExistingSavedData(worldAccess, registryAccess, WorldGenSettings.TYPE)
+        WorldGenSettings worldGenSettings = readExistingSavedData(worldAccess, dimension, registryAccess, WorldGenSettings.TYPE) // Paper - pass dimension
             .mapOrElse(
                 Function.identity(),
                 error -> {
@@ -170,9 +_,9 @@
     }
 
     public static <T extends SavedData> DataResult<T> readExistingSavedData(
-        LevelStorageSource.LevelStorageAccess access, HolderLookup.Provider registryAccess, SavedDataType<T> savedDataType
+        LevelStorageSource.LevelStorageAccess access, final ResourceKey<Level> dimension, HolderLookup.Provider registryAccess, SavedDataType<T> savedDataType
     ) {
-        Path dataLocation = savedDataType.id().withSuffix(".dat").resolveAgainst(access.getLevelPath(LevelResource.DATA));
+        Path dataLocation = savedDataType.id().withSuffix(".dat").resolveAgainst(access.getDimensionPath(dimension).resolve(LevelResource.DATA.id())); // Paper - use dimension scoped data (we 'demote' some data from global and add our own dimension scoped data)
 
         CompoundTag fileContents;
         try {
@@ -248,7 +_,7 @@
                 } catch (OutOfMemoryError e) {
                     MemoryReserve.release();
                     String detailedMessage = "Ran out of memory trying to read summary of world folder \"" + level.directoryName() + "\"";
-                    LOGGER.error(LogUtils.FATAL_MARKER, "{}", detailedMessage);
+                    LOGGER.error(LogUtils.FATAL_MARKER, detailedMessage);
                     OutOfMemoryError detailedException = new OutOfMemoryError("Ran out of memory reading level data");
                     detailedException.initCause(e);
                     CrashReport crashReport = CrashReport.forThrowable(detailedException, detailedMessage);
@@ -391,7 +_,7 @@
 
     public LevelStorageSource.LevelStorageAccess validateAndCreateAccess(String levelId) throws IOException, ContentValidationException {
         Path levelPath = this.getLevelPath(levelId);
-        List<ForbiddenSymlinkInfo> validationResults = this.worldDirValidator.validateDirectory(levelPath, true);
+        List<ForbiddenSymlinkInfo> validationResults = Boolean.getBoolean("paper.disableWorldSymlinkValidation") ? List.of() : this.worldDirValidator.validateDirectory(levelPath, true); // Paper - add skipping of symlinks scan
         if (!validationResults.isEmpty()) {
             throw new ContentValidationException(levelPath, validationResults);
         } else {
@@ -525,6 +_,19 @@
             }
         }
 
+        // Neo: For reading NeoForge's additional save data (mod list, etc.)
+        public void readAdditionalLevelSaveData(boolean fallback) {
+            checkLock();
+            Path path = fallback ? this.levelDirectory.oldDataFile() : this.levelDirectory.dataFile();
+            try {
+                var tag = readLightweightData(path);
+                if (tag instanceof CompoundTag compoundTag)
+                    net.neoforged.neoforge.common.CommonHooks.readAdditionalLevelSaveData(compoundTag, this.levelDirectory);
+            } catch (Exception e) {
+                LOGGER.error("Exception reading {}", path, e);
+            }
+        }
+
         public PlayerDataStorage createPlayerStorage() {
             this.checkLock();
             return new PlayerDataStorage(this, LevelStorageSource.this.fixerUpper);
@@ -554,10 +_,12 @@
             Dynamic<?> unfixedDataTag;
             try {
                 unfixedDataTag = this.getUnfixedDataTag(false);
+                this.readAdditionalLevelSaveData(false);
             } catch (IOException | NbtException | ReportedNbtException e) {
                 LevelStorageSource.LOGGER.warn("Failed to load world data from {}", this.levelDirectory.dataFile(), e);
                 LevelStorageSource.LOGGER.info("Attempting to use fallback {}", this.levelDirectory.oldDataFile());
                 unfixedDataTag = this.getUnfixedDataTag(true);
+                this.readAdditionalLevelSaveData(true);
                 this.restoreLevelDataFromOld();
             }
 
@@ -583,6 +_,7 @@
             CompoundTag dataTag = levelData.createTag(singleplayerUUID);
             CompoundTag root = new CompoundTag();
             root.put("Data", dataTag);
+            net.neoforged.neoforge.common.CommonHooks.writeAdditionalLevelSaveData(levelData, root);
             this.saveLevelData(root);
         }
 
@@ -609,6 +_,10 @@
 
         public Optional<Path> getIconFile() {
             return !this.lock.isValid() ? Optional.empty() : Optional.of(this.levelDirectory.iconFile());
+        }
+
+        public Path getWorldDir() {
+            return baseDir;
         }
 
         public void deleteLevel() throws IOException {
