--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -1,12 +_,12 @@
 package net.minecraft.world.level.storage;
 
+import com.mohistmc.youer.plugins.world.utils.ConfigByWorlds;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Sets;
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.DynamicOps;
 import com.mojang.serialization.Lifecycle;
-import com.mojang.serialization.DataResult.Error;
 import java.util.HashSet;
 import java.util.Set;
 import java.util.UUID;
@@ -16,6 +_,7 @@
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.UUIDUtil;
 import net.minecraft.nbt.CompoundTag;
@@ -24,7 +_,10 @@
 import net.minecraft.nbt.NbtUtils;
 import net.minecraft.nbt.StringTag;
 import net.minecraft.nbt.Tag;
+import net.minecraft.network.protocol.game.ClientboundChangeDifficultyPacket;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.GameType;
@@ -32,11 +_,15 @@
 import net.minecraft.world.level.LevelSettings;
 import net.minecraft.world.level.WorldDataConfiguration;
 import net.minecraft.world.level.border.WorldBorder;
+import net.minecraft.world.level.dimension.LevelStem;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
 import net.minecraft.world.level.levelgen.WorldGenSettings;
 import net.minecraft.world.level.levelgen.WorldOptions;
 import net.minecraft.world.level.timers.TimerCallbacks;
 import net.minecraft.world.level.timers.TimerQueue;
+import org.bukkit.Bukkit;
+import org.bukkit.event.weather.ThunderChangeEvent;
+import org.bukkit.event.weather.WeatherChangeEvent;
 import org.slf4j.Logger;
 
 public class PrimaryLevelData implements ServerLevelData, WorldData {
@@ -74,6 +_,28 @@
     private boolean wasModded;
     private final Set<String> removedFeatureFlags;
     private final TimerQueue<MinecraftServer> scheduledEvents;
+    private boolean confirmedExperimentalWarning = false;
+    // CraftBukkit start - Add world and pdc
+    public Registry<LevelStem> customDimensions;
+    private ServerLevel world;
+    protected Tag pdc;
+
+    public void setWorld(ServerLevel world) {
+        if (this.world != null) {
+            return;
+        }
+        this.world = world;
+        world.getWorld().readBukkitValues(pdc);
+        pdc = null;
+    }
+
+    // CraftBukkit start - Check if the name stored in NBT is the correct one
+    public void checkName(String name) {
+        if (!this.settings.levelName.equals(name)) {
+            this.settings.levelName = name;
+        }
+    }
+    // CraftBukkit end
 
     private PrimaryLevelData(
         @Nullable CompoundTag p_277888_,
@@ -169,7 +_,7 @@
         Dynamic<T> p_78531_, LevelSettings p_78535_, PrimaryLevelData.SpecialWorldProperty p_250651_, WorldOptions p_251864_, Lifecycle p_78538_
     ) {
         long i = p_78531_.get("Time").asLong(0L);
-        return new PrimaryLevelData(
+        PrimaryLevelData result = new PrimaryLevelData(
             p_78531_.get("Player").flatMap(CompoundTag.CODEC::parse).result().orElse(null),
             p_78531_.get("WasModded").asBoolean(false),
             new BlockPos(p_78531_.get("SpawnX").asInt(0), p_78531_.get("SpawnY").asInt(0), p_78531_.get("SpawnZ").asInt(0)),
@@ -200,7 +_,11 @@
             p_251864_,
             p_250651_,
             p_78538_
-        );
+        ).withConfirmedWarning(p_78538_ != Lifecycle.stable() && p_78531_.get("confirmedExperimentalSettings").asBoolean(false));
+        // Neo:
+        result.setDayTimeFraction(p_78531_.get("neoDayTimeFraction").asFloat(0f));
+        result.setDayTimePerTick(p_78531_.get("neoDayTimePerTick").asFloat(-1f));
+        return result;
     }
 
     @Override
@@ -273,6 +_,16 @@
         if (this.wanderingTraderId != null) {
             p_78547_.putUUID("WanderingTraderId", this.wanderingTraderId);
         }
+        p_78547_.putString("forgeLifecycle", net.neoforged.neoforge.common.CommonHooks.encodeLifecycle(this.settings.getLifecycle()));
+        p_78547_.putBoolean("confirmedExperimentalSettings", this.confirmedExperimentalWarning);
+        // Neo:
+        p_78547_.putFloat("neoDayTimeFraction", dayTimeFraction);
+        p_78547_.putFloat("neoDayTimePerTick", dayTimePerTick);
+
+        if (Bukkit.getServer() != null && world != null) {
+            p_78547_.putString("Bukkit.Version", Bukkit.getName() + "/" + Bukkit.getVersion() + "/" + Bukkit.getBukkitVersion()); // CraftBukkit
+            world.getWorld().storeBukkitValues(p_78547_); // CraftBukkit - add pdc
+        }
     }
 
     private static ListTag stringCollectionToTag(Set<String> p_277880_) {
@@ -350,6 +_,20 @@
 
     @Override
     public void setThundering(boolean p_78562_) {
+        // CraftBukkit start
+        if (this.thundering == p_78562_) {
+            return;
+        }
+
+        org.bukkit.World world = Bukkit.getWorld(getLevelName());
+        if (world != null) {
+            ThunderChangeEvent thunder = new ThunderChangeEvent(world, p_78562_);
+            Bukkit.getServer().getPluginManager().callEvent(thunder);
+            if (thunder.isCancelled()) {
+                return;
+            }
+        }
+        // CraftBukkit end
         this.thundering = p_78562_;
     }
 
@@ -370,6 +_,20 @@
 
     @Override
     public void setRaining(boolean p_78576_) {
+        // CraftBukkit start
+        if (this.raining == p_78576_) {
+            return;
+        }
+
+        org.bukkit.World world = Bukkit.getWorld(this.getLevelName());
+        if (world != null) {
+            WeatherChangeEvent weather = new WeatherChangeEvent(world, p_78576_);
+            Bukkit.getServer().getPluginManager().callEvent(weather);
+            if (weather.isCancelled()) {
+                return;
+            }
+        }
+        // CraftBukkit end
         this.raining = p_78576_;
     }
 
@@ -436,6 +_,12 @@
     @Override
     public void setDifficulty(Difficulty p_78521_) {
         this.settings = this.settings.withDifficulty(p_78521_);
+        // CraftBukkit start
+        ClientboundChangeDifficultyPacket packet = new ClientboundChangeDifficultyPacket(this.getDifficulty(), this.isDifficultyLocked());
+        for (ServerPlayer player : this.world.players()) {
+            player.connection.send(packet);
+        }
+        // CraftBukkit end
     }
 
     @Override
@@ -572,10 +_,44 @@
         return this.settings.copy();
     }
 
+    public boolean hasConfirmedExperimentalWarning() {
+        return this.confirmedExperimentalWarning;
+    }
+
+    public PrimaryLevelData withConfirmedWarning(boolean confirmedWarning) { // Builder-like to not patch ctor
+        this.confirmedExperimentalWarning = confirmedWarning;
+        return this;
+    }
+
     @Deprecated
     public static enum SpecialWorldProperty {
         NONE,
         FLAT,
         DEBUG;
+    }
+
+    // Neo: Variable day time code
+
+    private float dayTimeFraction = 0.0f;
+    private float dayTimePerTick = -1.0f;
+
+    @Override
+    public float getDayTimeFraction() {
+        return dayTimeFraction;
+    }
+
+    @Override
+    public float getDayTimePerTick() {
+        return dayTimePerTick;
+    }
+
+    @Override
+    public void setDayTimeFraction(float dayTimeFraction) {
+        this.dayTimeFraction = dayTimeFraction;
+    }
+
+    @Override
+    public void setDayTimePerTick(float dayTimePerTick) {
+        this.dayTimePerTick = dayTimePerTick;
     }
 }
