--- a/net/minecraft/client/renderer/chunk/ChunkSectionLayer.java
+++ b/net/minecraft/client/renderer/chunk/ChunkSectionLayer.java
@@ -10,19 +_,17 @@
 
 @OnlyIn(Dist.CLIENT)
 public enum ChunkSectionLayer {
-    SOLID(RenderPipelines.SOLID_TERRAIN, RenderPipelines.SOLID_TERRAIN_MULTIDRAW, 4194304, false),
-    CUTOUT(RenderPipelines.CUTOUT_TERRAIN, RenderPipelines.CUTOUT_TERRAIN_MULTIDRAW, 4194304, false),
-    TRANSLUCENT(RenderPipelines.TRANSLUCENT_TERRAIN, RenderPipelines.TRANSLUCENT_TERRAIN_MULTIDRAW, 786432, true);
+    SOLID(RenderPipelines.SOLID_TERRAIN, 4194304, false),
+    CUTOUT(RenderPipelines.CUTOUT_TERRAIN, 4194304, false),
+    TRANSLUCENT(RenderPipelines.TRANSLUCENT_TERRAIN, 786432, true);
 
     private final RenderPipeline pipeline;
-    private final RenderPipeline multiDrawPipeline;
     private final int bufferSize;
     private final boolean translucent;
     private final String label;
 
-    ChunkSectionLayer(RenderPipeline pipeline, RenderPipeline multiDrawPipeline, int bufferSize, boolean translucent) {
+    ChunkSectionLayer(RenderPipeline pipeline, int bufferSize, boolean translucent) {
         this.pipeline = pipeline;
-        this.multiDrawPipeline = multiDrawPipeline;
         this.bufferSize = bufferSize;
         this.translucent = translucent;
         this.label = this.toString().toLowerCase(Locale.ROOT);
@@ -36,8 +_,8 @@
         }
     }
 
-    public RenderPipeline pipeline(boolean multiDraw) {
-        return multiDraw ? this.multiDrawPipeline : this.pipeline;
+    public RenderPipeline pipeline() {
+        return this.pipeline;
     }
 
     public int bufferSize() {
