--- a/net/minecraft/client/renderer/chunk/ChunkSectionsToRender.java
+++ b/net/minecraft/client/renderer/chunk/ChunkSectionsToRender.java
@@ -4,16 +_,15 @@
 import com.mojang.renderpearl.api.buffers.GpuBuffer;
 import com.mojang.renderpearl.api.buffers.GpuBufferSlice;
 import com.mojang.renderpearl.api.commands.RenderPass;
-import com.mojang.renderpearl.api.device.DeviceLimits;
 import com.mojang.renderpearl.api.pipeline.IndexType;
 import com.mojang.renderpearl.api.pipeline.PrimitiveTopology;
 import com.mojang.renderpearl.api.pipeline.RenderPipeline;
 import com.mojang.renderpearl.api.textures.FilterMode;
 import com.mojang.renderpearl.api.textures.GpuSampler;
 import com.mojang.renderpearl.api.textures.GpuTextureView;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
 import java.util.EnumMap;
 import java.util.List;
-import java.util.Map;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.renderer.RenderPipelines;
@@ -24,49 +_,21 @@
 import org.jspecify.annotations.Nullable;
 
 @OnlyIn(Dist.CLIENT)
-public abstract class ChunkSectionsToRender {
-    private final int maxIndicesRequired;
-    private final GpuBufferSlice terrainTransformUBO;
-
-    private ChunkSectionsToRender(GpuBufferSlice terrainTransformUBO, int maxIndicesRequired) {
-        this.maxIndicesRequired = maxIndicesRequired;
-        this.terrainTransformUBO = terrainTransformUBO;
-    }
-
-    protected abstract void render(
-        final ChunkSectionLayer layer,
-        final RenderPass renderPass,
-        @Nullable GpuBuffer defaultIndexBuffer,
-        @Nullable IndexType defaultIndexType,
-        final @Nullable RenderPipeline renderPipelineOverride,
-        final @Nullable RenderPipeline renderPipelineOverrideMultidraw
-    );
-
-    public void renderGroup(ChunkSectionLayerGroup group, RenderPass renderPass, GpuSampler sampler, GpuTextureView atlas, boolean renderWireframeTerrain) {
+public record ChunkSectionsToRender(
+    GpuTextureView textureView,
+    EnumMap<ChunkSectionLayer, Int2ObjectOpenHashMap<List<RenderPass.Draw<GpuBufferSlice[]>>>> drawGroupsPerLayer,
+    int maxIndicesRequired,
+    GpuBufferSlice[] chunkSectionInfos
+) {
+    public void renderGroup(ChunkSectionLayerGroup group, RenderPass renderPass, GpuSampler sampler, boolean renderWireframeTerrain) {
         GameRenderer gameRenderer = Minecraft.getInstance().gameRenderer;
         GpuTextureView lightmap = gameRenderer.lightmap();
-        this.renderLayers(
-            group.layers(),
-            sampler,
-            renderPass,
-            atlas,
-            lightmap,
-            renderWireframeTerrain ? RenderPipelines.WIREFRAME : null,
-            renderWireframeTerrain ? RenderPipelines.WIREFRAME_MULTIDRAW : null
-        );
+        this.renderLayers(group.layers(), sampler, renderPass, lightmap, renderWireframeTerrain ? RenderPipelines.WIREFRAME : null, true);
     }
 
-    public void renderOit(GpuSampler sampler, OitStage stage, OitRenderPassProvider.Parameters params, GpuTextureView atlas, GpuTextureView lightmap) {
+    public void renderOit(GpuSampler sampler, OitStage stage, OitRenderPassProvider.Parameters params, GpuTextureView lightmap) {
         try (RenderPass renderPass = OitRenderPassProvider.createRenderPass(stage, () -> "Terrain", params)) {
-            this.renderLayers(
-                ChunkSectionLayerGroup.TRANSLUCENT.layers(),
-                sampler,
-                renderPass,
-                atlas,
-                lightmap,
-                RenderPipelines.OIT_TERRAIN.getPipeline(stage),
-                RenderPipelines.OIT_TERRAIN_MULTIDRAW.getPipeline(stage)
-            );
+            this.renderLayers(ChunkSectionLayerGroup.TRANSLUCENT.layers(), sampler, renderPass, lightmap, RenderPipelines.OIT_TERRAIN.getPipeline(stage), false);
         }
     }
 
@@ -74,115 +_,32 @@
         ChunkSectionLayer[] layers,
         GpuSampler sampler,
         RenderPass renderPass,
-        GpuTextureView atlas,
         GpuTextureView lightmap,
         @Nullable RenderPipeline renderPipelineOverride,
-        @Nullable RenderPipeline renderPipelineOverrideMultidraw
+        boolean shouldReverseTranslucent
     ) {
         RenderSystem.AutoStorageIndexBuffer autoIndices = RenderSystem.getSequentialBuffer(PrimitiveTopology.QUADS);
         GpuBuffer defaultIndexBuffer = this.maxIndicesRequired == 0 ? null : autoIndices.getBuffer();
         IndexType defaultIndexType = this.maxIndicesRequired == 0 ? null : autoIndices.type();
-        renderPass.setUniform("TerrainUniform", this.terrainTransformUBO);
-        renderPass.setUniform("Sampler0", atlas, sampler);
-        renderPass.setUniform("Sampler2", lightmap, RenderSystem.getSamplerCache().getClampToEdge(FilterMode.LINEAR));
+        renderPass.bindTexture("Sampler0", this.textureView, sampler);
+        renderPass.bindTexture("Sampler2", lightmap, RenderSystem.getSamplerCache().getClampToEdge(FilterMode.LINEAR));
 
         for (ChunkSectionLayer layer : layers) {
             renderPass.pushDebugGroup(() -> "Terrain layer: " + layer.label());
-            this.render(layer, renderPass, defaultIndexBuffer, defaultIndexType, renderPipelineOverride, renderPipelineOverrideMultidraw);
-            renderPass.popDebugGroup();
-        }
-    }
-
-    public static final class DrawIndirect extends ChunkSectionsToRender {
-        private final EnumMap<ChunkSectionLayer, List<ChunkSectionsToRender.GpuMultiDrawIndexedIndirect>> drawGroupsPerLayer;
-        private final GpuBufferSlice chunkSectionInfos;
-
-        public DrawIndirect(
-            GpuBufferSlice terrainTransformUBO,
-            EnumMap<ChunkSectionLayer, List<ChunkSectionsToRender.GpuMultiDrawIndexedIndirect>> drawGroupsPerLayer,
-            int maxIndicesRequired,
-            GpuBufferSlice chunkSectionInfos
-        ) {
-            super(terrainTransformUBO, maxIndicesRequired);
-            this.drawGroupsPerLayer = drawGroupsPerLayer;
-            this.chunkSectionInfos = chunkSectionInfos;
-        }
-
-        @Override
-        protected void render(
-            ChunkSectionLayer layer,
-            RenderPass renderPass,
-            @Nullable GpuBuffer defaultIndexBuffer,
-            @Nullable IndexType defaultIndexType,
-            @Nullable RenderPipeline renderPipelineOverride,
-            @Nullable RenderPipeline renderPipelineOverrideMultidraw
-        ) {
-            renderPass.setPipeline(
-                RenderSystem.getCompiledPipeline(renderPipelineOverrideMultidraw != null ? renderPipelineOverrideMultidraw : layer.pipeline(true))
-            );
-            DeviceLimits deviceLimits = RenderSystem.getDevice().getDeviceInfo().limits();
-            int maxDrawIndirectDrawCount = deviceLimits.maxDrawIndirectDrawCount();
-            List<ChunkSectionsToRender.GpuMultiDrawIndexedIndirect> drawGroups = this.drawGroupsPerLayer.get(layer);
-            if (!drawGroups.isEmpty()) {
-                renderPass.setVertexBuffer(1, this.chunkSectionInfos);
-            }
-
-            for (ChunkSectionsToRender.GpuMultiDrawIndexedIndirect indirectDraw : drawGroups) {
-                if (indirectDraw.drawCount() > 0) {
-                    renderPass.setVertexBuffer(0, indirectDraw.vertexBuffer);
-                    IndexType indexType = indirectDraw.indexType() == null ? defaultIndexType : indirectDraw.indexType();
-                    renderPass.setIndexBuffer(indirectDraw.indexBuffer() == null ? defaultIndexBuffer : indirectDraw.indexBuffer().buffer(), indexType);
-                    GpuBuffer buffer = indirectDraw.indirectCommandBuffer.buffer();
-                    long startOffset = indirectDraw.indirectCommandBuffer().offset();
-                    int remainingDrawCount = indirectDraw.drawCount();
-
-                    while (remainingDrawCount > 0) {
-                        int passDrawCount = Integer.min(remainingDrawCount, maxDrawIndirectDrawCount);
-                        long length = passDrawCount * 20L;
-                        GpuBufferSlice passSlice = buffer.slice(startOffset, length);
-                        renderPass.drawIndexedIndirect(passSlice, passDrawCount);
-                        remainingDrawCount -= passDrawCount;
-                        startOffset += length;
+            renderPass.setPipeline(RenderSystem.getCompiledPipeline(renderPipelineOverride != null ? renderPipelineOverride : layer.pipeline()));
+            Int2ObjectOpenHashMap<List<RenderPass.Draw<GpuBufferSlice[]>>> drawGroup = this.drawGroupsPerLayer.get(layer);
+
+            for (List<RenderPass.Draw<GpuBufferSlice[]>> draws : drawGroup.values()) {
+                if (!draws.isEmpty()) {
+                    if (shouldReverseTranslucent && layer == ChunkSectionLayer.TRANSLUCENT) {
+                        draws = draws.reversed();
                     }
+
+                    renderPass.drawMultipleIndexed(draws, defaultIndexBuffer, defaultIndexType, List.of("ChunkSection"), this.chunkSectionInfos);
                 }
             }
-        }
-    }
-
-    public static final class DrawSeparate extends ChunkSectionsToRender {
-        private final Map<ChunkSectionLayer, List<RenderPass.Draw<GpuBufferSlice[]>>> drawsPerLayer;
-        private final GpuBufferSlice[] chunkSectionInfos;
-
-        public DrawSeparate(
-            GpuBufferSlice terrainTransformUBO,
-            Map<ChunkSectionLayer, List<RenderPass.Draw<GpuBufferSlice[]>>> drawsPerLayer,
-            int maxIndicesRequired,
-            GpuBufferSlice[] chunkSectionInfos
-        ) {
-            super(terrainTransformUBO, maxIndicesRequired);
-            this.drawsPerLayer = drawsPerLayer;
-            this.chunkSectionInfos = chunkSectionInfos;
-        }
-
-        @Override
-        protected void render(
-            ChunkSectionLayer layer,
-            RenderPass renderPass,
-            @Nullable GpuBuffer defaultIndexBuffer,
-            @Nullable IndexType defaultIndexType,
-            @Nullable RenderPipeline renderPipelineOverride,
-            @Nullable RenderPipeline renderPipelineOverrideMultidraw
-        ) {
-            renderPass.setPipeline(RenderSystem.getCompiledPipeline(renderPipelineOverride != null ? renderPipelineOverride : layer.pipeline(false)));
-            List<RenderPass.Draw<GpuBufferSlice[]>> draws = this.drawsPerLayer.get(layer);
-            if (!draws.isEmpty()) {
-                renderPass.drawMultipleIndexed(draws, defaultIndexBuffer, defaultIndexType, List.of("ChunkSection"), this.chunkSectionInfos);
-            }
-        }
-    }
-
-    public record GpuMultiDrawIndexedIndirect(
-        GpuBufferSlice vertexBuffer, @Nullable GpuBufferSlice indexBuffer, @Nullable IndexType indexType, GpuBufferSlice indirectCommandBuffer, int drawCount
-    ) {
+
+            renderPass.popDebugGroup();
+        }
     }
 }
