--- a/net/minecraft/client/renderer/chunk/RenderSectionRegion.java
+++ b/net/minecraft/client/renderer/chunk/RenderSectionRegion.java
@@ -45,6 +_,16 @@
     }
 
     @Override
+    public @Nullable BlockState getBlockStateIfLoaded(BlockPos pos) {
+        return null;
+    }
+
+    @Override
+    public @Nullable FluidState getFluidIfLoaded(BlockPos pos) {
+        return null;
+    }
+
+    @Override
     public FluidState getFluidState(BlockPos pos) {
         return this.getSection(
                 SectionPos.blockToSectionCoord(pos.getX()), SectionPos.blockToSectionCoord(pos.getY()), SectionPos.blockToSectionCoord(pos.getZ())
