--- a/net/minecraft/client/renderer/chunk/SectionRenderDispatcher.java
+++ b/net/minecraft/client/renderer/chunk/SectionRenderDispatcher.java
@@ -39,8 +_,6 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.IRenderableSection;
-
 @OnlyIn(Dist.CLIENT)
 public class SectionRenderDispatcher {
     public static final int NEARBY_SECTION_DISTANCE_IN_BLOCKS = 32;
@@ -203,7 +_,7 @@
         return this.bufferPool.getFreeBufferCount();
     }
 
-    public class RenderSection implements RotatingSectionStorage.Value, IRenderableSection {
+    public class RenderSection implements RotatingSectionStorage.Value {
         public final int index;
         public final AtomicReference<SectionMesh> sectionMesh = new AtomicReference<>(CompiledSectionMesh.UNCOMPILED);
         private SectionRenderDispatcher.RenderSection.@Nullable CompileTask lastCompileTask;
