--- a/net/minecraft/world/item/enchantment/effects/AllOf.java
+++ b/net/minecraft/world/item/enchantment/effects/AllOf.java
@@ -28,7 +_,7 @@
         return new AllOf.ValueEffects(List.of(effects));
     }
 
-    record EntityEffects(List<EnchantmentEntityEffect> effects) implements EnchantmentEntityEffect {
+    public record EntityEffects(List<EnchantmentEntityEffect> effects) implements EnchantmentEntityEffect {
         public static final MapCodec<AllOf.EntityEffects> CODEC = AllOf.codec(
             EnchantmentEntityEffect.CODEC, AllOf.EntityEffects::new, AllOf.EntityEffects::effects
         );
@@ -46,7 +_,7 @@
         }
     }
 
-    record LocationBasedEffects(List<EnchantmentLocationBasedEffect> effects) implements EnchantmentLocationBasedEffect {
+    public record LocationBasedEffects(List<EnchantmentLocationBasedEffect> effects) implements EnchantmentLocationBasedEffect {
         public static final MapCodec<AllOf.LocationBasedEffects> CODEC = AllOf.codec(
             EnchantmentLocationBasedEffect.CODEC, AllOf.LocationBasedEffects::new, AllOf.LocationBasedEffects::effects
         );
@@ -71,7 +_,7 @@
         }
     }
 
-    record ValueEffects(List<EnchantmentValueEffect> effects) implements EnchantmentValueEffect {
+    public record ValueEffects(List<EnchantmentValueEffect> effects) implements EnchantmentValueEffect {
         public static final MapCodec<AllOf.ValueEffects> CODEC = AllOf.codec(EnchantmentValueEffect.CODEC, AllOf.ValueEffects::new, AllOf.ValueEffects::effects);
 
         @Override
