--- a/net/minecraft/world/item/enchantment/effects/ApplyMobEffect.java
+++ b/net/minecraft/world/item/enchantment/effects/ApplyMobEffect.java
@@ -41,6 +_,7 @@
             if (optional.isPresent()) {
                 int i = Math.round(Mth.randomBetween(randomsource, this.minDuration.calculate(p_346112_), this.maxDuration.calculate(p_346112_)) * 20.0F);
                 int j = Math.max(0, Math.round(Mth.randomBetween(randomsource, this.minAmplifier.calculate(p_346112_), this.maxAmplifier.calculate(p_346112_))));
+                livingentity.addEffectCause(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK);
                 livingentity.addEffect(new MobEffectInstance(optional.get(), i, j));
             }
         }
