--- a/net/minecraft/world/item/enchantment/effects/ChangeItemDamage.java
+++ b/net/minecraft/world/item/enchantment/effects/ChangeItemDamage.java
@@ -4,7 +_,6 @@
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.core.component.DataComponents;
 import net.minecraft.server.level.ServerLevel;
-import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.enchantment.EnchantedItemInUse;
@@ -20,9 +_,9 @@
     public void apply(ServerLevel serverLevel, int enchantmentLevel, EnchantedItemInUse item, Entity entity, Vec3 position) {
         ItemStack itemStack = item.itemStack();
         if (itemStack.has(DataComponents.MAX_DAMAGE) && itemStack.has(DataComponents.DAMAGE)) {
-            ServerPlayer player = item.owner() instanceof ServerPlayer sp ? sp : null;
+            //ServerPlayer player = item.owner() instanceof ServerPlayer sp ? sp : null; // Paper - EntityDamageItemEvent - always pass in entity
             int change = (int)this.amount.calculate(enchantmentLevel);
-            itemStack.hurtAndBreak(change, serverLevel, player, item.onBreak());
+            itemStack.hurtAndBreak(change, serverLevel, item.owner(), item.onBreak()); // Paper - EntityDamageItemEvent - always pass in entity
         }
     }
 
