--- a/net/minecraft/world/item/enchantment/effects/DamageItem.java
+++ b/net/minecraft/world/item/enchantment/effects/DamageItem.java
@@ -18,7 +_,7 @@
     @Override
     public void apply(ServerLevel p_345496_, int p_344926_, EnchantedItemInUse p_345827_, Entity p_345408_, Vec3 p_345110_) {
         ServerPlayer serverplayer = p_345827_.owner() instanceof ServerPlayer serverplayer1 ? serverplayer1 : null;
-        p_345827_.itemStack().hurtAndBreak((int)this.amount.calculate(p_344926_), p_345496_, serverplayer, p_345827_.onBreak());
+        p_345827_.itemStack().hurtAndBreak((int)this.amount.calculate(p_344926_), p_345496_, p_345827_.owner(), p_345827_.onBreak());
     }
 
     @Override
