--- a/net/minecraft/world/item/enchantment/effects/ReplaceBlock.java
+++ b/net/minecraft/world/item/enchantment/effects/ReplaceBlock.java
@@ -2,7 +_,6 @@
 
 import com.mojang.serialization.MapCodec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import java.util.Optional;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
@@ -14,6 +_,7 @@
 import net.minecraft.world.level.levelgen.blockpredicates.BlockPredicate;
 import net.minecraft.world.level.levelgen.feature.stateproviders.BlockStateProvider;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public record ReplaceBlock(Vec3i offset, Optional<BlockPredicate> predicate, BlockStateProvider blockState, Optional<Holder<GameEvent>> triggerGameEvent)
     implements EnchantmentEntityEffect {
@@ -30,8 +_,9 @@
     @Override
     public void apply(ServerLevel p_345756_, int p_346215_, EnchantedItemInUse p_345202_, Entity p_346293_, Vec3 p_345339_) {
         BlockPos blockpos = BlockPos.containing(p_345339_).offset(this.offset);
+        p_345756_.callEvent(true);
         if (this.predicate.map(p_345193_ -> p_345193_.test(p_345756_, blockpos)).orElse(true)
-            && p_345756_.setBlockAndUpdate(blockpos, this.blockState.getState(p_346293_.getRandom(), blockpos))) {
+            && (p_345756_.setBlockAndUpdate(blockpos, this.blockState.getState(p_346293_.getRandom(), blockpos)) || CraftEventFactory.handleBlockFormEvent(p_345756_, blockpos, this.blockState.getState(p_346293_.getRandom(), blockpos), p_346293_))) {
             this.triggerGameEvent.ifPresent(p_347354_ -> p_345756_.gameEvent(p_346293_, (Holder<GameEvent>)p_347354_, blockpos));
         }
     }
