--- a/net/minecraft/world/level/biome/BiomeManager.java
+++ b/net/minecraft/world/level/biome/BiomeManager.java
@@ -103,7 +_,6 @@
     }
 
     private static double getFiddle(long rval) {
-        double uniform = Math.floorMod(rval >> 24, 1024) / 1024.0;
-        return (uniform - 0.5) * 0.9;
+        return (double)(((rval >> 24) & (1024 - 1)) - (1024/2)) * (0.9 / 1024.0); // Paper - avoid floorMod, fp division, and fp subtraction
     }
 }
