--- a/net/minecraft/world/level/biome/CheckerboardColumnBiomeSource.java
+++ b/net/minecraft/world/level/biome/CheckerboardColumnBiomeSource.java
@@ -7,7 +_,7 @@
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderSet;
 
-public class CheckerboardColumnBiomeSource extends BiomeSource implements BiomeResolver {
+public class CheckerboardColumnBiomeSource extends BiomeSource {
     public static final MapCodec<CheckerboardColumnBiomeSource> CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(
                 Biome.LIST_CODEC.fieldOf("biomes").forGetter(s -> s.allowedBiomes), Codec.intRange(0, 62).optionalFieldOf("scale", 2).forGetter(s -> s.size)
@@ -30,17 +_,12 @@
     }
 
     @Override
-    public BiomeResolver createResolver(Climate.Sampler sampler) {
-        return this;
-    }
-
-    @Override
-    protected MapCodec<CheckerboardColumnBiomeSource> codec() {
+    protected MapCodec<? extends BiomeSource> codec() {
         return CODEC;
     }
 
     @Override
-    public Holder<Biome> getNoiseBiome(int quartX, int quartY, int quartZ) {
+    public Holder<Biome> getNoiseBiome(int quartX, int quartY, int quartZ, Climate.Sampler sampler) {
         return this.allowedBiomes.get(Math.floorMod((quartX >> this.bitShift) + (quartZ >> this.bitShift), this.allowedBiomes.size()));
     }
 }
