--- a/net/minecraft/world/level/biome/MobSpawnSettings.java
+++ b/net/minecraft/world/level/biome/MobSpawnSettings.java
@@ -44,6 +_,8 @@
     private final float creatureGenerationProbability;
     private final Map<MobCategory, WeightedList<MobSpawnSettings.SpawnerData>> spawners;
     private final Map<EntityType<?>, MobSpawnSettings.MobSpawnCost> mobSpawnCosts;
+    private final java.util.Set<MobCategory> typesView;
+    private final java.util.Set<EntityType<?>> costView;
 
     private MobSpawnSettings(
         float creatureGenerationProbability,
@@ -53,24 +_,58 @@
         this.creatureGenerationProbability = creatureGenerationProbability;
         this.spawners = ImmutableMap.copyOf(spawners);
         this.mobSpawnCosts = ImmutableMap.copyOf(mobSpawnCosts);
+        this.typesView = java.util.Collections.unmodifiableSet(this.spawners.keySet());
+        this.costView = java.util.Collections.unmodifiableSet(this.mobSpawnCosts.keySet());
     }
 
     public WeightedList<MobSpawnSettings.SpawnerData> getMobs(MobCategory category) {
         return this.spawners.getOrDefault(category, EMPTY_MOB_LIST);
     }
 
+    public java.util.Set<MobCategory> getSpawnerTypes() {
+         return this.typesView;
+    }
+
     public MobSpawnSettings.@Nullable MobSpawnCost getMobSpawnCost(EntityType<?> type) {
         return this.mobSpawnCosts.get(type);
     }
 
+    public java.util.Set<EntityType<?>> getEntityTypes() {
+         return this.costView;
+    }
+
     public float getCreatureProbability() {
         return this.creatureGenerationProbability;
     }
 
     public static class Builder {
-        protected final Map<MobCategory, WeightedList.Builder<MobSpawnSettings.SpawnerData>> spawners = Util.makeEnumMap(
-            MobCategory.class, c -> WeightedList.builder()
+        // Paper start - Perf: keep track of data in a pair set to give O(1) contains calls - we have to hook removals incase plugins mess with it
+        public static class MobListBuilder<E> extends WeightedList.Builder<E> {
+            @Override
+            public WeightedList<E> build() {
+                return new WeightedSpawnerDataList<>(this.result);
+            }
+        }
+
+        public static class WeightedSpawnerDataList<E> extends WeightedList<E> {
+            private final java.util.Set<E> spawnerDataSet = new java.util.HashSet<>();
+
+            public WeightedSpawnerDataList(final java.util.List<? extends net.minecraft.util.random.Weighted<E>> items) {
+                super(items);
+                for (final net.minecraft.util.random.Weighted<E> item : items) {
+                    this.spawnerDataSet.add(item.value());
+                }
+            }
+
+            @Override
+            public boolean contains(final E element) {
+                return this.spawnerDataSet.contains(element);
+            }
+        }
+        public final Map<MobCategory, WeightedList.Builder<MobSpawnSettings.SpawnerData>> spawners = Util.makeEnumMap(
+                MobCategory.class, c -> new MobListBuilder<>()
         );
+        // Paper end - Perf: keep track of data in a pair set to give O(1) contains calls
         protected final Map<EntityType<?>, MobSpawnSettings.MobSpawnCost> mobSpawnCosts = Maps.newLinkedHashMap();
         protected float creatureGenerationProbability = 0.1F;
 
