--- a/net/minecraft/world/level/biome/OverworldBiomeBuilder.java
+++ b/net/minecraft/world/level/biome/OverworldBiomeBuilder.java
@@ -18,12 +_,11 @@
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.util.CubicSpline;
 import net.minecraft.util.VisibleForDebug;
+import net.minecraft.world.level.levelgen.DensityFunction;
+import net.minecraft.world.level.levelgen.DensityFunctions;
 import net.minecraft.world.level.levelgen.NoiseRouterData;
 import net.minecraft.world.level.levelgen.OverworldFunctionSet;
 import net.minecraft.world.level.levelgen.SpawnTargetPoint;
-import net.minecraft.world.level.levelgen.densityfunction.DensityFunction;
-import net.minecraft.world.level.levelgen.densityfunction.DensityFunctions;
-import net.minecraft.world.level.levelgen.densityfunction.op.SplineFunction;
 
 public final class OverworldBiomeBuilder {
     private static final float VALLEY_SIZE = 0.05F;
@@ -165,13 +_,13 @@
             .add(Registries.NOISE, NoiseData::bootstrap)
             .build(RegistryAccess.fromRegistryOfRegistries(BuiltInRegistries.REGISTRY));
         HolderGetter<DensityFunction> densityFunctions = builtIns.lookupOrThrow(Registries.DENSITY_FUNCTION);
-        SplineFunction.Coordinate continents = new SplineFunction.Coordinate(
+        DensityFunctions.Spline.Coordinate continents = new DensityFunctions.Spline.Coordinate(
             new DensityFunctions.HolderHolder(densityFunctions.getOrThrow(NoiseRouterData.OVERWORLD_FUNCTIONS.continents()))
         );
-        SplineFunction.Coordinate erosion = new SplineFunction.Coordinate(
+        DensityFunctions.Spline.Coordinate erosion = new DensityFunctions.Spline.Coordinate(
             new DensityFunctions.HolderHolder(densityFunctions.getOrThrow(NoiseRouterData.OVERWORLD_FUNCTIONS.erosion()))
         );
-        SplineFunction.Coordinate ridges = new SplineFunction.Coordinate(
+        DensityFunctions.Spline.Coordinate ridges = new DensityFunctions.Spline.Coordinate(
             new DensityFunctions.HolderHolder(densityFunctions.getOrThrow(NoiseRouterData.RIDGES_FOLDED))
         );
         biomes.accept(
@@ -1056,7 +_,7 @@
 
     public static DensityFunction deepDarkRegion(DensityFunction erosion, DensityFunction depth) {
         return DensityFunctions.min(
-            DensityFunctions.sub(DensityFunctions.constant(-0.225F), erosion), DensityFunctions.max(depth.sub(0.9F), DensityFunctions.constant(0.0F))
+            DensityFunctions.sub(DensityFunctions.constant(-0.225F), erosion), DensityFunctions.max(depth.sub(0.9F), DensityFunctions.constant(0.0))
         );
     }
 
