--- a/net/minecraft/world/level/levelgen/structure/placement/AbstractSpreadingStructurePlacement.java
+++ b/net/minecraft/world/level/levelgen/structure/placement/AbstractSpreadingStructurePlacement.java
@@ -20,11 +_,11 @@
 
 public abstract class AbstractSpreadingStructurePlacement implements StructurePlacement {
     private static final int HIGHLY_ARBITRARY_RANDOM_SALT = 10387320;
-    public final Vec3i locateOffset;
-    public final AbstractSpreadingStructurePlacement.FrequencyReductionMethod frequencyReductionMethod;
-    public final float frequency;
-    public final int salt;
-    public final Optional<AbstractSpreadingStructurePlacement.ExclusionZone> exclusionZone;
+    private final Vec3i locateOffset;
+    private final AbstractSpreadingStructurePlacement.FrequencyReductionMethod frequencyReductionMethod;
+    private final float frequency;
+    private final int salt;
+    private final Optional<AbstractSpreadingStructurePlacement.ExclusionZone> exclusionZone;
 
     protected static <S extends AbstractSpreadingStructurePlacement> P5<Mu<S>, Vec3i, AbstractSpreadingStructurePlacement.FrequencyReductionMethod, Float, Integer, Optional<AbstractSpreadingStructurePlacement.ExclusionZone>> placementCodec(
         Instance<S> i
