--- a/net/minecraft/world/level/levelgen/structure/placement/AbstractSpreadingStructurePlacement.java
+++ b/net/minecraft/world/level/levelgen/structure/placement/AbstractSpreadingStructurePlacement.java
@@ -81,7 +_,16 @@
     }
 
     @Override
-    public boolean isStructureChunk(ChunkGeneratorStructureState state, int sourceX, int sourceZ) {
+    public boolean isStructureChunk(ChunkGeneratorStructureState state, int sourceX, int sourceZ, final net.minecraft.resources.@org.jspecify.annotations.Nullable ResourceKey<net.minecraft.world.level.levelgen.structure.StructureSet> structureSetKey) {
+        Integer saltOverride = null;
+        if (structureSetKey != null) {
+            if (structureSetKey == net.minecraft.world.level.levelgen.structure.BuiltinStructureSets.MINESHAFTS) {
+                saltOverride = state.conf.mineshaftSeed;
+            } else if (structureSetKey == net.minecraft.world.level.levelgen.structure.BuiltinStructureSets.BURIED_TREASURES) {
+                saltOverride = state.conf.buriedTreasureSeed;
+            }
+        }
+        // Paper end - Add missing structure set seed configs
         return this.isPlacementChunk(state, sourceX, sourceZ)
             && this.applyAdditionalChunkRestrictions(sourceX, sourceZ, state.getLevelSeed())
             && this.applyInteractionsWithOtherStructures(state, sourceX, sourceZ);
