--- a/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EndCrystal.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.entity.EntityTypes;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
+import net.minecraft.world.level.ExplosionDamageCalculator;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BaseFireBlock;
 import net.minecraft.world.level.dimension.end.EnderDragonFight;
@@ -20,6 +_,12 @@
 import net.minecraft.world.level.storage.ValueOutput;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+// CraftBukkit end
+
 public class EndCrystal extends Entity {
     private static final EntityDataAccessor<Optional<BlockPos>> DATA_BEAM_TARGET = SynchedEntityData.defineId(
         EndCrystal.class, EntityDataSerializers.OPTIONAL_BLOCK_POS
@@ -27,6 +_,13 @@
     private static final EntityDataAccessor<Boolean> DATA_SHOW_BOTTOM = SynchedEntityData.defineId(EndCrystal.class, EntityDataSerializers.BOOLEAN);
     private static final boolean DEFAULT_SHOW_BOTTOM = true;
     public int time;
+    public boolean generatedByDragonFight = false; // Paper - Fix invulnerable end crystals
+    // Purpur start - Phantoms attracted to crystals and crystals shoot phantoms
+    private net.minecraft.world.entity.monster.Phantom targetPhantom;
+    private int phantomBeamTicks = 0;
+    private int phantomDamageCooldown = 0;
+    private int idleCooldown = 0;
+    // Purpur end - Phantoms attracted to crystals and crystals shoot phantoms
 
     public EndCrystal(EntityType<? extends EndCrystal> type, Level level) {
         super(type, level);
@@ -39,6 +_,24 @@
         this.setPos(x, y, z);
     }
 
+    // Purpur start - End crystal explosion options
+    public boolean shouldExplode() {
+        return showsBottom() ? level().purpurConfig.basedEndCrystalExplode : level().purpurConfig.baselessEndCrystalExplode;
+    }
+
+    public float getExplosionPower() {
+        return (float) (showsBottom() ? level().purpurConfig.basedEndCrystalExplosionPower : level().purpurConfig.baselessEndCrystalExplosionPower);
+    }
+
+    public boolean hasExplosionFire() {
+        return showsBottom() ? level().purpurConfig.basedEndCrystalExplosionFire : level().purpurConfig.baselessEndCrystalExplosionFire;
+    }
+
+    public Level.ExplosionInteraction getExplosionEffect() {
+        return showsBottom() ? level().purpurConfig.basedEndCrystalExplosionEffect : level().purpurConfig.baselessEndCrystalExplosionEffect;
+    }
+    // Purpur end - End crystal explosion options
+
     @Override
     protected Entity.MovementEmission getMovementEmission() {
         return Entity.MovementEmission.NONE;
@@ -55,24 +_,84 @@
         this.time++;
         this.applyEffectsFromBlocks();
         this.handlePortal();
-        if (this.level() instanceof ServerLevel serverLevel) {
+        if (this.level() instanceof ServerLevel) {
             BlockPos pos = this.blockPosition();
-            if (serverLevel.getDragonFight() != null && serverLevel.getBlockState(pos).isAir()) {
-                serverLevel.setBlockAndUpdate(pos, BaseFireBlock.getState(serverLevel, pos));
-            }
-        }
+            if (((ServerLevel)this.level()).getDragonFight() != null && this.level().getBlockState(pos).isAir()) {
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBlockIgniteEvent(this.level(), pos, this).isCancelled()) { // Paper
+                    this.level().setBlockAndUpdate(pos, BaseFireBlock.getState(this.level(), pos));
+                } // Paper
+            }
+        }
+
+        // Paper start - Fix invulnerable end crystals
+        if (this.level().paperConfig().unsupportedSettings.fixInvulnerableEndCrystalExploit && this.generatedByDragonFight && this.isInvulnerable()) {
+            if (!java.util.Objects.equals(((ServerLevel) this.level()).uuid, this.originWorld)
+                    || ((ServerLevel) this.level()).getDragonFight() == null
+                    || ((ServerLevel) this.level()).getDragonFight().respawnStage == null
+                    || ((ServerLevel) this.level()).getDragonFight().respawnStage.ordinal() > net.minecraft.world.level.dimension.end.DragonRespawnStage.SUMMONING_DRAGON.ordinal()) {
+                this.setInvulnerable(false);
+                this.setBeamTarget(null);
+            }
+        }
+        // Paper end - Fix invulnerable end crystals
+        if (this.level().purpurConfig.endCrystalCramming > 0 && this.level().getEntitiesOfClass(EndCrystal.class, getBoundingBox()).size() > this.level().purpurConfig.endCrystalCramming) this.hurt(this.damageSources().cramming(), 6.0F); // Purpur - End Crystal Cramming
+        // Purpur start - Phantoms attracted to crystals and crystals shoot phantoms
+        if (level().purpurConfig.phantomAttackedByCrystalRadius <= 0 || --idleCooldown > 0) {
+            return; // on cooldown
+        }
+
+        if (targetPhantom == null) {
+            for (net.minecraft.world.entity.monster.Phantom phantom : level().getEntitiesOfClass(net.minecraft.world.entity.monster.Phantom.class, getBoundingBox().inflate(level().purpurConfig.phantomAttackedByCrystalRadius))) {
+                if (phantom.hasLineOfSight(this)) {
+                    attackPhantom(phantom);
+                    break;
+                }
+            }
+        } else {
+            setBeamTarget(new BlockPos(targetPhantom).offset(0, -2, 0));
+            if (--phantomBeamTicks > 0 && targetPhantom.isAlive()) {
+                phantomDamageCooldown--;
+                if (targetPhantom.hasLineOfSight(this)) {
+                    if (phantomDamageCooldown <= 0) {
+                        phantomDamageCooldown = 20;
+                        targetPhantom.hurt(targetPhantom.damageSources().indirectMagic(this, this), level().purpurConfig.phantomAttackedByCrystalDamage);
+                    }
+                } else {
+                    forgetPhantom(); // no longer in sight
+                }
+            } else {
+                forgetPhantom(); // attacked long enough
+            }
+        }
+    }
+
+    private void attackPhantom(net.minecraft.world.entity.monster.Phantom phantom) {
+        phantomDamageCooldown = 0;
+        phantomBeamTicks = 60;
+        targetPhantom = phantom;
+    }
+
+    private void forgetPhantom() {
+        targetPhantom = null;
+        setBeamTarget(null);
+        phantomBeamTicks = 0;
+        phantomDamageCooldown = 0;
+        idleCooldown = 60;
+        // Purpur end - Phantoms attracted to crystals and crystals shoot phantoms
     }
 
     @Override
     protected void addAdditionalSaveData(ValueOutput output) {
         output.storeNullable("beam_target", BlockPos.CODEC, this.getBeamTarget());
         output.putBoolean("ShowBottom", this.showsBottom());
+        output.putBoolean("Paper.GeneratedByDragonFight", this.generatedByDragonFight); // Paper - Fix invulnerable end crystals
     }
 
     @Override
     protected void readAdditionalSaveData(ValueInput input) {
         this.setBeamTarget(input.read("beam_target", BlockPos.CODEC).orElse(null));
         this.setShowBottom(input.getBooleanOr("ShowBottom", true));
+        this.generatedByDragonFight = input.getBooleanOr("Paper.GeneratedByDragonFight", false); // Paper - Fix invulnerable end crystals
     }
 
     @Override
@@ -89,23 +_,38 @@
     public final boolean hurtServer(ServerLevel level, DamageSource source, float damage) {
         if (this.isInvulnerableToBase(source)) {
             return false;
-        }
-
-        if (source.getEntity() instanceof EnderDragon) {
+        } else if (source.getEntity() instanceof EnderDragon) {
             return false;
-        }
-
-        if (!this.isRemoved()) {
-            this.remove(Entity.RemovalReason.KILLED);
-            if (!source.is(DamageTypeTags.IS_EXPLOSION)) {
-                DamageSource damageSource = source.getEntity() != null ? this.damageSources().explosion(this, source.getEntity()) : null;
-                level.explode(this, damageSource, null, this.getX(), this.getY(), this.getZ(), 6.0F, false, Level.ExplosionInteraction.BLOCK);
+        } else {
+            if (!this.isRemoved()) {
+                // CraftBukkit start - All non-living entities need this
+                if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, source, damage, false)) {
+                    return false;
+                }
+                // CraftBukkit end
+                if (!source.is(DamageTypeTags.IS_EXPLOSION)) {
+                    if (shouldExplode()) {// Purpur - End crystal explosion options
+                        DamageSource damagesource1 = source.getEntity() != null ? this.damageSources().explosion(this, source.getEntity()) : null;
+
+                        // CraftBukkit start
+                        org.bukkit.event.entity.ExplosionPrimeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callExplosionPrimeEvent(this, getExplosionPower(), hasExplosionFire()); // Purpur - End crystal explosion options
+                        if (event.isCancelled()) {
+                            return false;
+                        }
+
+                        this.remove(Entity.RemovalReason.KILLED, EntityRemoveEvent.Cause.EXPLODE); // CraftBukkit - add Bukkit remove cause
+                        level.explode(this, damagesource1, null, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), getExplosionEffect()); // Purpur - End crystal explosion options
+                    } else this.unsetRemoved(); // Purpur - End crystal explosion options
+                } else {
+                    this.remove(Entity.RemovalReason.KILLED, EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause
+                }
+                // CraftBukkit end
+
+                this.onDestroyedBy(level, source);
             }
 
-            this.onDestroyedBy(level, source);
+            return true;
         }
-
-        return true;
     }
 
     @Override
