--- a/net/minecraft/world/entity/monster/spider/Spider.java
+++ b/net/minecraft/world/entity/monster/spider/Spider.java
@@ -167,6 +_,7 @@
         if (groupData instanceof Spider.SpiderEffectsGroupData spiderEffectsGroupData) {
             Holder<MobEffect> effect = spiderEffectsGroupData.effect;
             if (effect != null) {
+                this.addEffectCause(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.SPIDER_SPAWN); // CraftBukkit
                 this.addEffect(new MobEffectInstance(effect, -1));
             }
         }
