--- a/net/minecraft/world/attribute/modifier/AttributeModifier.java
+++ b/net/minecraft/world/attribute/modifier/AttributeModifier.java
@@ -7,8 +_,6 @@
 import net.minecraft.world.attribute.EnvironmentAttribute;
 import net.minecraft.world.attribute.LerpFunction;
 import net.minecraft.world.level.biome.MobSpawnSettings;
-import org.joml.Vector3fc;
-import org.joml.Vector4fc;
 
 public interface AttributeModifier<Subject, Argument> {
     Map<AttributeModifier.OperationId, AttributeModifier<Boolean, ?>> BOOLEAN_LIBRARY = Map.of(
@@ -39,29 +_,29 @@
         AttributeModifier.OperationId.MAXIMUM,
         FloatModifier.MAXIMUM
     );
-    Map<AttributeModifier.OperationId, AttributeModifier<Vector3fc, ?>> RGB_COLOR_LIBRARY = Map.of(
+    Map<AttributeModifier.OperationId, AttributeModifier<Integer, ?>> RGB_COLOR_LIBRARY = Map.of(
         AttributeModifier.OperationId.ALPHA_BLEND,
-        ColorModifier.ALPHA_BLEND_RGB,
+        ColorModifier.ALPHA_BLEND,
         AttributeModifier.OperationId.ADD,
-        ColorModifier.ADD_RGB,
+        ColorModifier.ADD,
         AttributeModifier.OperationId.SUBTRACT,
-        ColorModifier.SUBTRACT_RGB,
+        ColorModifier.SUBTRACT,
         AttributeModifier.OperationId.MULTIPLY,
         ColorModifier.MULTIPLY_RGB,
         AttributeModifier.OperationId.BLEND_TO_GRAY,
-        ColorModifier.BLEND_TO_GRAY_RGB
+        ColorModifier.BLEND_TO_GRAY
     );
-    Map<AttributeModifier.OperationId, AttributeModifier<Vector4fc, ?>> ARGB_COLOR_LIBRARY = Map.of(
+    Map<AttributeModifier.OperationId, AttributeModifier<Integer, ?>> ARGB_COLOR_LIBRARY = Map.of(
         AttributeModifier.OperationId.ALPHA_BLEND,
-        ColorModifier.ALPHA_BLEND_ARGB,
+        ColorModifier.ALPHA_BLEND,
         AttributeModifier.OperationId.ADD,
-        ColorModifier.ADD_ARGB,
+        ColorModifier.ADD,
         AttributeModifier.OperationId.SUBTRACT,
-        ColorModifier.SUBTRACT_ARGB,
+        ColorModifier.SUBTRACT,
         AttributeModifier.OperationId.MULTIPLY,
         ColorModifier.MULTIPLY_ARGB,
         AttributeModifier.OperationId.BLEND_TO_GRAY,
-        ColorModifier.BLEND_TO_GRAY_ARGB
+        ColorModifier.BLEND_TO_GRAY
     );
     Map<AttributeModifier.OperationId, AttributeModifier<Integer, ?>> INTEGER_LIBRARY = Map.of(
         AttributeModifier.OperationId.ADD,
