--- a/net/minecraft/world/attribute/modifier/ColorModifier.java
+++ b/net/minecraft/world/attribute/modifier/ColorModifier.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.attribute.modifier;
 
+import com.mojang.datafixers.util.Either;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
 import net.minecraft.util.ARGB;
@@ -7,61 +_,55 @@
 import net.minecraft.util.Mth;
 import net.minecraft.world.attribute.EnvironmentAttribute;
 import net.minecraft.world.attribute.LerpFunction;
-import org.joml.Vector3fc;
-import org.joml.Vector4fc;
-
-public interface ColorModifier<Subject, Argument> extends AttributeModifier<Subject, Argument> {
-    ColorModifier.ArgbModifier<Vector3fc> ALPHA_BLEND_RGB = ARGB::alphaBlend;
-    ColorModifier.ArgbModifier<Vector4fc> ALPHA_BLEND_ARGB = ARGB::alphaBlend;
-    ColorModifier.RgbModifier<Vector3fc> ADD_RGB = ARGB::addRgb;
-    ColorModifier.RgbModifier<Vector3fc> SUBTRACT_RGB = ARGB::subtractRgb;
-    ColorModifier.RgbModifier<Vector3fc> MULTIPLY_RGB = ARGB::multiply;
-    ColorModifier.RgbModifier<Vector4fc> ADD_ARGB = ARGB::addRgb;
-    ColorModifier.RgbModifier<Vector4fc> SUBTRACT_ARGB = ARGB::subtractRgb;
-    ColorModifier.ArgbModifier<Vector4fc> MULTIPLY_ARGB = ARGB::multiply;
-    ColorModifier<Vector3fc, ColorModifier.BlendToGray> BLEND_TO_GRAY_RGB = new ColorModifier<Vector3fc, ColorModifier.BlendToGray>() {
-        public Vector3fc apply(Vector3fc subject, ColorModifier.BlendToGray argument) {
-            Vector3fc multipliedGreyscale = ARGB.scaleRGB(ARGB.greyscale(subject), argument.brightness);
-            return ARGB.srgbLerp(argument.factor, subject, multipliedGreyscale);
-        }
-
-        @Override
-        public Codec<ColorModifier.BlendToGray> argumentCodec(EnvironmentAttribute<Vector3fc> type) {
-            return ColorModifier.BlendToGray.CODEC;
-        }
-
-        @Override
-        public LerpFunction<ColorModifier.BlendToGray> argumentKeyframeLerp(EnvironmentAttribute<Vector3fc> type) {
-            return ColorModifier.BlendToGray::lerp;
+
+public interface ColorModifier<Argument> extends AttributeModifier<Integer, Argument> {
+    ColorModifier<Integer> ALPHA_BLEND = new ColorModifier<Integer>() {
+        public Integer apply(Integer subject, Integer argument) {
+            return ARGB.alphaBlend(subject, argument);
+        }
+
+        @Override
+        public Codec<Integer> argumentCodec(EnvironmentAttribute<Integer> type) {
+            return ExtraCodecs.STRING_ARGB_COLOR;
+        }
+
+        @Override
+        public LerpFunction<Integer> argumentKeyframeLerp(EnvironmentAttribute<Integer> type) {
+            return LerpFunction.ofColor();
         }
     };
-    ColorModifier<Vector4fc, ColorModifier.BlendToGray> BLEND_TO_GRAY_ARGB = new ColorModifier<Vector4fc, ColorModifier.BlendToGray>() {
-        public Vector4fc apply(Vector4fc subject, ColorModifier.BlendToGray argument) {
-            Vector4fc multipliedGreyscale = ARGB.scaleRGB(ARGB.greyscale(subject), argument.brightness);
+    ColorModifier<Integer> ADD = (RgbModifier) ARGB::addRgb;
+    ColorModifier<Integer> SUBTRACT = (RgbModifier) ARGB::subtractRgb;
+    ColorModifier<Integer> MULTIPLY_RGB = (RgbModifier) ARGB::multiply;
+    ColorModifier<Integer> MULTIPLY_ARGB = (ArgbModifier) ARGB::multiply;
+    ColorModifier<ColorModifier.BlendToGray> BLEND_TO_GRAY = new ColorModifier<ColorModifier.BlendToGray>() {
+        public Integer apply(Integer subject, ColorModifier.BlendToGray argument) {
+            int multipliedGreyscale = ARGB.scaleRGB(ARGB.greyscale(subject), argument.brightness);
             return ARGB.srgbLerp(argument.factor, subject, multipliedGreyscale);
         }
 
         @Override
-        public Codec<ColorModifier.BlendToGray> argumentCodec(EnvironmentAttribute<Vector4fc> type) {
+        public Codec<ColorModifier.BlendToGray> argumentCodec(EnvironmentAttribute<Integer> type) {
             return ColorModifier.BlendToGray.CODEC;
         }
 
         @Override
-        public LerpFunction<ColorModifier.BlendToGray> argumentKeyframeLerp(EnvironmentAttribute<Vector4fc> type) {
-            return ColorModifier.BlendToGray::lerp;
+        public LerpFunction<ColorModifier.BlendToGray> argumentKeyframeLerp(EnvironmentAttribute<Integer> type) {
+            return (alpha, from, to) -> new ColorModifier.BlendToGray(Mth.lerp(alpha, from.brightness, to.brightness), Mth.lerp(alpha, from.factor, to.factor));
         }
     };
 
     @FunctionalInterface
-    interface ArgbModifier<Subject> extends ColorModifier<Subject, Vector4fc> {
+    interface ArgbModifier extends ColorModifier<Integer> {
         @Override
-        default Codec<Vector4fc> argumentCodec(EnvironmentAttribute<Subject> type) {
-            return ExtraCodecs.STRING_ARGB_VEC4_COLOR;
+        default Codec<Integer> argumentCodec(EnvironmentAttribute<Integer> type) {
+            return Codec.either(ExtraCodecs.STRING_ARGB_COLOR, ExtraCodecs.RGB_COLOR_CODEC)
+                .xmap(Either::unwrap, color -> ARGB.alpha(color) == 255 ? Either.right(color) : Either.left(color));
         }
 
         @Override
-        default LerpFunction<Vector4fc> argumentKeyframeLerp(EnvironmentAttribute<Subject> type) {
-            return LerpFunction.ofColorVec4();
+        default LerpFunction<Integer> argumentKeyframeLerp(EnvironmentAttribute<Integer> type) {
+            return LerpFunction.ofColor();
         }
     }
 
@@ -73,21 +_,18 @@
                 )
                 .apply(i, ColorModifier.BlendToGray::new)
         );
-
-        public static ColorModifier.BlendToGray lerp(float alpha, ColorModifier.BlendToGray from, ColorModifier.BlendToGray to) {
-            return new ColorModifier.BlendToGray(Mth.lerp(alpha, from.brightness, to.brightness), Mth.lerp(alpha, from.factor, to.factor));
-        }
     }
 
-    interface RgbModifier<Subject> extends ColorModifier<Subject, Vector3fc> {
+    @FunctionalInterface
+    interface RgbModifier extends ColorModifier<Integer> {
         @Override
-        default Codec<Vector3fc> argumentCodec(EnvironmentAttribute<Subject> type) {
-            return ExtraCodecs.STRING_RGB_VEC3_COLOR;
+        default Codec<Integer> argumentCodec(EnvironmentAttribute<Integer> type) {
+            return ExtraCodecs.STRING_RGB_COLOR;
         }
 
         @Override
-        default LerpFunction<Vector3fc> argumentKeyframeLerp(EnvironmentAttribute<Subject> type) {
-            return LerpFunction.ofColorVec3();
+        default LerpFunction<Integer> argumentKeyframeLerp(EnvironmentAttribute<Integer> type) {
+            return LerpFunction.ofColor();
         }
     }
 }
