--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -18,7 +_,6 @@
 import net.minecraft.world.damagesource.DamageSource;
 import net.minecraft.world.entity.AnimationState;
 import net.minecraft.world.entity.Entity;
-import net.minecraft.world.entity.EntityEvent;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.ai.Brain;
@@ -93,6 +_,14 @@
         this.xpReward = 0;
     }
 
+    // Purpur start - Configurable entity base attributes
+    @Override
+    public void initAttributes() {
+        this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(this.level().purpurConfig.creakingMaxHealth);
+        this.getAttribute(Attributes.SCALE).setBaseValue(this.level().purpurConfig.creakingScale);
+    }
+    // Purpur end - Configurable entity base attributes
+
     public void setTransient(BlockPos pos) {
         this.setHomePos(pos);
         this.setPathfindingMalus(PathType.DAMAGING, 8.0F);
@@ -193,9 +_,9 @@
     }
 
     @Override
-    public void push(double xa, double ya, double za) {
+    public void push(final double xa, final double ya, final double za, @Nullable Entity pushingEntity) { // Paper - add push source entity param
         if (this.canMove()) {
-            super.push(xa, ya, za);
+            super.push(xa, ya, za, pushingEntity); // Paper - add push source entity param
         }
     }
 
@@ -320,7 +_,7 @@
         }
 
         this.makeSound(this.getDeathSound());
-        this.remove(Entity.RemovalReason.DISCARDED);
+        this.remove(Entity.RemovalReason.DISCARDED, null); // CraftBukkit - add Bukkit remove cause
     }
 
     public void creakingDeathEffects(DamageSource source) {
@@ -330,7 +_,7 @@
     }
 
     @Override
-    public void handleEntityEvent(@EntityEvent.Value byte id) {
+    public void handleEntityEvent(byte id) {
         if (id == 66) {
             this.invulnerabilityAnimationRemainingTicks = 8;
             this.playHurtSound(this.damageSources().generic());
@@ -448,9 +_,9 @@
     }
 
     @Override
-    public void knockback(double power, double xd, double zd, DamageSource source, float damage, boolean comesFromEffect) {
+    public void knockback(final double power, final double xd, final double zd, final DamageSource source, final float damage, final boolean comesFromEffect, @Nullable final Entity attacker, final io.papermc.paper.event.entity.EntityKnockbackEvent.Cause cause) { // Paper - knockback events
         if (this.canMove()) {
-            super.knockback(power, xd, zd, source, damage, comesFromEffect);
+            super.knockback(power, xd, zd, source, damage, comesFromEffect, attacker, cause); // Paper - knockback events
         }
     }
 
@@ -469,7 +_,8 @@
             for (Player player : players) {
                 if (this.canAttack(player) && !this.isAlliedTo(player)) {
                     hasPotentialTarget = true;
-                    if ((!active || LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM.test(player))
+                    // Neo: provide entity being looked at to disguise check
+                    if ((!active || LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET.test(player, this))
                         && this.isLookingAtMe(
                             player, 0.5, false, true, this.getEyeY(), this.getY() + 0.5 * this.getScale(), (this.getEyeY() + this.getY()) / 2.0
                         )) {
