--- a/net/minecraft/world/entity/monster/creaking/Creaking.java
+++ b/net/minecraft/world/entity/monster/creaking/Creaking.java
@@ -431,17 +_,17 @@
     }
 
     @Override
-    protected SoundEvent getAmbientSound() {
+    public SoundEvent getAmbientSound() {
         return this.isActive() ? null : SoundEvents.CREAKING_AMBIENT;
     }
 
     @Override
-    protected SoundEvent getHurtSound(DamageSource p_380378_) {
+    public SoundEvent getHurtSound(DamageSource p_380378_) {
         return this.isHeartBound() ? SoundEvents.CREAKING_SWAY : super.getHurtSound(p_380378_);
     }
 
     @Override
-    protected SoundEvent getDeathSound() {
+    public SoundEvent getDeathSound() {
         return SoundEvents.CREAKING_DEATH;
     }
 
@@ -478,7 +_,8 @@
             for (Player player : list) {
                 if (this.canAttack(player) && !this.isAlliedTo(player)) {
                     flag1 = true;
-                    if ((!flag || LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM.test(player))
+                    // Neo: provide entity being looked at to disguise check
+                    if ((!flag || LivingEntity.PLAYER_NOT_WEARING_DISGUISE_ITEM_FOR_TARGET.test(player, this))
                         && this.isLookingAtMe(
                             player, 0.5, false, true, new double[]{this.getEyeY(), this.getY() + 0.5 * this.getScale(), (this.getEyeY() + this.getY()) / 2.0}
                         )) {
