--- a/com/mojang/renderpearl/backend/vulkan/checkpoints/AmdCheckpointExtension.java
+++ b/com/mojang/renderpearl/backend/vulkan/checkpoints/AmdCheckpointExtension.java
@@ -4,7 +_,6 @@
 import com.mojang.renderpearl.backend.vulkan.VulkanDevice;
 import com.mojang.renderpearl.backend.vulkan.VulkanGpuBuffer;
 import com.mojang.renderpearl.backend.vulkan.VulkanQueue;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.util.ArrayList;
 import java.util.List;
 import net.neoforged.api.distmarker.Dist;
@@ -42,7 +_,7 @@
         }
     }
 
-    private static class AmdCheckpointStorage extends AbstractCheckpointStorage implements UncheckedAutoCloseable {
+    private static class AmdCheckpointStorage extends AbstractCheckpointStorage implements AutoCloseable {
         private final VulkanGpuBuffer buffer;
         private final GpuBufferSlice.MappedView mappedView;
 
