--- a/com/mojang/renderpearl/backend/vulkan/checkpoints/CheckpointExtension.java
+++ b/com/mojang/renderpearl/backend/vulkan/checkpoints/CheckpointExtension.java
@@ -2,7 +_,6 @@
 
 import com.mojang.renderpearl.backend.vulkan.VulkanDevice;
 import com.mojang.renderpearl.backend.vulkan.VulkanQueue;
-import com.mojang.renderpearl.util.UncheckedAutoCloseable;
 import java.util.List;
 import java.util.function.Supplier;
 import net.neoforged.api.distmarker.Dist;
@@ -10,10 +_,13 @@
 import org.lwjgl.vulkan.VkCommandBuffer;
 
 @OnlyIn(Dist.CLIENT)
-public interface CheckpointExtension extends UncheckedAutoCloseable {
+public interface CheckpointExtension extends AutoCloseable {
     CheckpointExtension.CheckpointStorage createStorage(VulkanDevice device, VulkanQueue queue, int maxFramesInFlight);
 
     List<CheckpointExtension.QueueCheckpoints> retrieveCheckpoints(boolean isDeviceLost);
+
+    @Override
+    void close();
 
     interface CheckpointStorage {
         void rotate();
