--- a/net/minecraft/util/worldupdate/WorldUpgrader.java
+++ b/net/minecraft/util/worldupdate/WorldUpgrader.java
@@ -111,6 +_,7 @@
         LOGGER.info("Upgrading blocks");
         new WorldUpgrader.ChunkUpgrader().upgrade();
         this.overworldDataStorage.save();
+        net.neoforged.neoforge.common.IOUtilities.waitUntilIOWorkerComplete();
         i = Util.getMillis() - i;
         LOGGER.info("World optimizaton finished after {} seconds", i / 1000L);
         this.finished = true;
@@ -348,7 +_,7 @@
                 int i = ChunkStorage.getVersion(compoundtag);
                 ChunkGenerator chunkgenerator = WorldUpgrader.this.dimensions.getOrThrow(Registries.levelToLevelStem(p_321488_)).generator();
                 CompoundTag compoundtag1 = p_321581_.upgradeChunkTag(
-                    p_321488_, () -> WorldUpgrader.this.overworldDataStorage, compoundtag, chunkgenerator.getTypeNameForDataFixer()
+                    p_321488_, () -> WorldUpgrader.this.overworldDataStorage, compoundtag, chunkgenerator.getTypeNameForDataFixer(), p_321524_, null // CraftBukkit
                 );
                 ChunkPos chunkpos = new ChunkPos(compoundtag1.getInt("xPos"), compoundtag1.getInt("zPos"));
                 if (!chunkpos.equals(p_321524_)) {
