--- a/net/minecraft/client/resources/model/EquipmentAssetManager.java
+++ b/net/minecraft/client/resources/model/EquipmentAssetManager.java
@@ -1,6 +_,5 @@
 package net.minecraft.client.resources.model;
 
-import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.stream.Collectors;
@@ -17,7 +_,7 @@
 
 @OnlyIn(Dist.CLIENT)
 public class EquipmentAssetManager extends SimpleJsonResourceReloadListener<EquipmentClientInfo> {
-    public static final EquipmentClientInfo MISSING = new EquipmentClientInfo(Map.of(), List.of());
+    public static final EquipmentClientInfo MISSING = new EquipmentClientInfo(Map.of(), Map.of());
     private static final FileToIdConverter ASSET_LISTER = FileToIdConverter.json("equipment");
     private Map<ResourceKey<EquipmentAsset>, EquipmentClientInfo> equipmentAssets = Map.of();
 
