--- a/net/minecraft/client/resources/model/ModelBaker.java
+++ b/net/minecraft/client/resources/model/ModelBaker.java
@@ -8,11 +_,8 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.joml.Vector3fc;
 
-import net.neoforged.neoforge.client.extensions.ModelBakerExtension;
-import net.neoforged.neoforge.client.extensions.ModelBakerInternerExtension;
-
 @OnlyIn(Dist.CLIENT)
-public interface ModelBaker extends ModelBakerExtension {
+public interface ModelBaker {
     ResolvedModel getModel(Identifier location);
 
     BlockStateModelPart missingBlockModelPart();
@@ -23,7 +_,7 @@
 
     <T> T compute(ModelBaker.SharedOperationKey<T> key);
 
-    interface Interner extends ModelBakerInternerExtension {
+    interface Interner {
         Vector3fc vector(Vector3fc vector);
 
         BakedQuad.MaterialInfo materialInfo(BakedQuad.MaterialInfo material);
