--- a/net/minecraft/client/resources/model/UnbakedModel.java
+++ b/net/minecraft/client/resources/model/UnbakedModel.java
@@ -8,10 +_,8 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.extensions.UnbakedModelExtension;
-
 @OnlyIn(Dist.CLIENT)
-public interface UnbakedModel extends UnbakedModelExtension {
+public interface UnbakedModel {
     String PARTICLE_TEXTURE_REFERENCE = "particle";
 
     default @Nullable Boolean ambientOcclusion() {
