--- a/net/minecraft/world/entity/ai/behavior/BehaviorFollowAdult.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorFollowAdult.java
@@ -7,6 +7,13 @@
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.ai.memory.MemoryTarget;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
+// CraftBukkit end
+
 public class BehaviorFollowAdult {
 
     public BehaviorFollowAdult() {}
@@ -27,6 +34,17 @@
                         EntityLiving entityliving1 = (EntityLiving) behaviorbuilder_b.get(memoryaccessor);
 
                         if (entityliving.closerThan(entityliving1, (double) (uniformint.getMaxValue() + 1)) && !entityliving.closerThan(entityliving1, (double) uniformint.getMinValue())) {
+                            // CraftBukkit start
+                            EntityTargetLivingEntityEvent event = CraftEventFactory.callEntityTargetLivingEvent(entityliving, entityliving1, EntityTargetEvent.TargetReason.FOLLOW_LEADER);
+                            if (event.isCancelled()) {
+                                return false;
+                            }
+                            if (event.getTarget() == null) {
+                                memoryaccessor.erase();
+                                return true;
+                            }
+                            entityliving1 = ((CraftLivingEntity) event.getTarget()).getHandle();
+                            // CraftBukkit end
                             MemoryTarget memorytarget = new MemoryTarget(new BehaviorPositionEntity(entityliving1, flag, flag), (Float) function.apply(entityliving), uniformint.getMinValue() - 1);
 
                             memoryaccessor1.set(new BehaviorPositionEntity(entityliving1, true, flag));
