--- a/net/minecraft/server/players/CachedUserNameToIdResolver.java
+++ b/net/minecraft/server/players/CachedUserNameToIdResolver.java
@@ -122,7 +122,7 @@
             cachedusernametoidresolver_a.setLastAccess(this.getNextOperation());
             optional = Optional.of(cachedusernametoidresolver_a.nameAndId());
         } else {
-            Optional<NameAndId> optional1 = this.lookupGameProfile(this.profileRepository, s1);
+            Optional<NameAndId> optional1 = this.lookupGameProfile(this.profileRepository, s); // CraftBukkit - use correct case for offline players
 
             if (optional1.isPresent()) {
                 optional = Optional.of(this.addInternal((NameAndId) optional1.get()).nameAndId());
@@ -168,7 +168,7 @@
             DateFormat dateformat = createDateFormat();
 
             jsonarray.forEach((jsonelement) -> {
-                Optional optional = readGameProfile(jsonelement, dateformat);
+                Optional<CachedUserNameToIdResolver.a> optional = readGameProfile(jsonelement, dateformat); // CraftBukkit - decompile error
 
                 Objects.requireNonNull(list);
                 optional.ifPresent(list::add);
