--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -104,6 +104,27 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import com.mojang.authlib.GameProfile;
+import java.util.stream.Collectors;
+import net.minecraft.server.dedicated.DedicatedServer;
+import net.minecraft.server.level.ClientInformation;
+import net.minecraft.server.network.ServerLoginPacketListenerImpl;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.util.CraftChatMessage;
+import org.bukkit.craftbukkit.util.CraftLocation;
+import org.bukkit.event.entity.EntityRemoveEvent;
+import org.bukkit.event.player.PlayerChangedWorldEvent;
+import org.bukkit.event.player.PlayerJoinEvent;
+import org.bukkit.event.player.PlayerLoginEvent;
+import org.bukkit.event.player.PlayerQuitEvent;
+import org.bukkit.event.player.PlayerRespawnEvent.RespawnReason;
+import org.bukkit.event.player.PlayerSpawnChangeEvent;
+// CraftBukkit end
+
 public abstract class PlayerList {
 
     public static final File USERBANLIST_FILE = new File("banned-players.json");
@@ -116,14 +137,16 @@
     private static final int SEND_PLAYER_INFO_INTERVAL = 600;
     private static final SimpleDateFormat BAN_DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z", Locale.ROOT);
     private final MinecraftServer server;
-    public final List<ServerPlayer> players = Lists.newArrayList();
+    public final List<ServerPlayer> players = new java.util.concurrent.CopyOnWriteArrayList(); // CraftBukkit - ArrayList -> CopyOnWriteArrayList: Iterator safety
     private final Map<UUID, ServerPlayer> playersByUUID = Maps.newHashMap();
     private final UserBanList bans;
     private final IpBanList ipBans;
     private final ServerOpList ops;
     private final UserWhiteList whitelist;
-    private final Map<UUID, ServerStatsCounter> stats = Maps.newHashMap();
-    private final Map<UUID, PlayerAdvancements> advancements = Maps.newHashMap();
+    // CraftBukkit start
+    // private final Map<UUID, ServerStatisticManager> stats = Maps.newHashMap();
+    // private final Map<UUID, AdvancementDataPlayer> advancements = Maps.newHashMap();
+    // CraftBukkit end
     public final PlayerDataStorage playerIo;
     private final LayeredRegistryAccess<RegistryLayer> registries;
     private int viewDistance;
@@ -131,7 +154,13 @@
     private boolean allowCommandsForAllPlayers;
     private int sendAllPlayerInfoIn;
 
+    // CraftBukkit start
+    private CraftServer cserver;
+
     public PlayerList(MinecraftServer minecraftserver, LayeredRegistryAccess<RegistryLayer> layeredregistryaccess, PlayerDataStorage playerdatastorage, NotificationService notificationservice) {
+        this.cserver = minecraftserver.server = new CraftServer((DedicatedServer) minecraftserver, this);
+        minecraftserver.console = org.bukkit.craftbukkit.command.ColouredConsoleSender.getInstance();
+        // CraftBukkit end
         this.server = minecraftserver;
         this.registries = layeredregistryaccess;
         this.playerIo = playerdatastorage;
@@ -146,12 +175,19 @@
         UserNameToIdResolver usernametoidresolver = this.server.services().nameToIdCache();
         Optional<NameAndId> optional = usernametoidresolver.get(nameandid.id());
         String s = (String) optional.map(NameAndId::name).orElse(nameandid.name());
+        // CraftBukkit start - Better rename detection
+        String lastKnownName = serverplayer.getBukkitEntity().getLastKnownName();
+        if (lastKnownName != null) {
+            s = lastKnownName;
+        }
+        // CraftBukkit end
 
         usernametoidresolver.add(nameandid);
         ServerLevel serverlevel = serverplayer.level();
         String s1 = connection.getLoggableAddress(this.server.logIPs());
 
-        PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ({}, {}, {})", new Object[]{serverplayer.getPlainTextName(), s1, serverplayer.getId(), serverplayer.getX(), serverplayer.getY(), serverplayer.getZ()});
+        // CraftBukkit - Moved message to after join
+        // PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ({}, {}, {})", new Object[]{serverplayer.getPlainTextName(), s1, serverplayer.getId(), serverplayer.getX(), serverplayer.getY(), serverplayer.getZ()});
         LevelData leveldata = serverlevel.getLevelData();
         ServerGamePacketListenerImpl servergamepacketlistenerimpl = new ServerGamePacketListenerImpl(this.server, connection, serverplayer, commonlistenercookie);
 
@@ -163,6 +199,7 @@
         boolean flag2 = (Boolean) gamerules.get(GameRules.LIMITED_CRAFTING);
 
         servergamepacketlistenerimpl.send(new ClientboundLoginPacket(serverplayer.getId(), leveldata.isHardcore(), this.server.levelKeys(), this.getMaxPlayers(), this.getViewDistance(), this.getSimulationDistance(), flag1, !flag, flag2, serverplayer.createCommonSpawnInfo(serverlevel), this.server.enforceSecureProfile()));
+        serverplayer.getBukkitEntity().sendSupportedChannels(); // CraftBukkit
         servergamepacketlistenerimpl.send(new ClientboundChangeDifficultyPacket(leveldata.getDifficulty(), leveldata.isDifficultyLocked()));
         servergamepacketlistenerimpl.send(new ClientboundPlayerAbilitiesPacket(serverplayer.getAbilities()));
         servergamepacketlistenerimpl.send(new ClientboundSetHeldSlotPacket(serverplayer.getInventory().getSelectedSlot()));
@@ -182,7 +219,10 @@
             mutablecomponent = Component.translatable("multiplayer.player.joined.renamed", serverplayer.getDisplayName(), s);
         }
 
-        this.broadcastSystemMessage(mutablecomponent.withStyle(ChatFormatting.YELLOW), false);
+        // CraftBukkit start
+        mutablecomponent.withStyle(ChatFormatting.YELLOW);
+        String joinMessage = CraftChatMessage.fromComponent(mutablecomponent);
+
         servergamepacketlistenerimpl.teleport(serverplayer.getX(), serverplayer.getY(), serverplayer.getZ(), serverplayer.getYRot(), serverplayer.getXRot());
         ServerStatus serverstatus = this.server.getStatus();
 
@@ -190,17 +230,70 @@
             serverplayer.sendServerStatus(serverstatus);
         }
 
-        serverplayer.connection.send(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(this.players));
+        // entityplayer.connection.send(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(this.players)); // CraftBukkit - replaced with loop below
         this.players.add(serverplayer);
         this.playersByUUID.put(serverplayer.getUUID(), serverplayer);
-        this.broadcastAll(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(serverplayer)));
+        // this.broadcastAll(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(entityplayer))); // CraftBukkit - replaced with loop below
+
+        // CraftBukkit start
+        CraftPlayer bukkitPlayer = serverplayer.getBukkitEntity();
+
+        // Ensure that player inventory is populated with its viewer
+        serverplayer.containerMenu.transferTo(serverplayer.containerMenu, bukkitPlayer);
+
+        PlayerJoinEvent playerJoinEvent = new PlayerJoinEvent(bukkitPlayer, joinMessage);
+        cserver.getPluginManager().callEvent(playerJoinEvent);
+
+        if (!serverplayer.connection.isAcceptingMessages()) {
+            return;
+        }
+
+        joinMessage = playerJoinEvent.getJoinMessage();
+
+        if (joinMessage != null && joinMessage.length() > 0) {
+            for (Component line : org.bukkit.craftbukkit.util.CraftChatMessage.fromString(joinMessage)) {
+                server.getPlayerList().broadcastSystemMessage(line, false);
+            }
+        }
+        // CraftBukkit end
+
+        // CraftBukkit start - sendAll above replaced with this loop
+        ClientboundPlayerInfoUpdatePacket packet = ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(serverplayer));
+
+        for (int i = 0; i < this.players.size(); ++i) {
+            ServerPlayer serverplayer1 = (ServerPlayer) this.players.get(i);
+
+            if (serverplayer1.getBukkitEntity().canSee(bukkitPlayer)) {
+                serverplayer1.connection.send(packet);
+            }
+
+            if (!bukkitPlayer.canSee(serverplayer1.getBukkitEntity())) {
+                continue;
+            }
+
+            serverplayer.connection.send(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(serverplayer1)));
+        }
+        serverplayer.sentListPacket = true;
+        // CraftBukkit end
+
+        serverplayer.refreshEntityData(serverplayer); // CraftBukkit - BungeeCord#2321, send complete data to self on spawn
+
         this.sendLevelInfo(serverplayer, serverlevel);
-        serverlevel.addNewPlayer(serverplayer);
-        this.server.getCustomBossEvents().onPlayerConnect(serverplayer);
+
+        // CraftBukkit start - Only add if the player wasn't moved in the event
+        if (serverplayer.level() == serverlevel && !serverlevel.players().contains(serverplayer)) {
+            serverlevel.addNewPlayer(serverplayer);
+            this.server.getCustomBossEvents().onPlayerConnect(serverplayer);
+        }
+
+        serverlevel = serverplayer.level(); // CraftBukkit - Update in case join event changed it
+        // CraftBukkit end
         this.sendActivePlayerEffects(serverplayer);
         serverplayer.initInventoryMenu();
         this.server.notificationManager().playerJoined(serverplayer);
         servergamepacketlistenerimpl.resumeFlushing();
+        // CraftBukkit - Moved from above, added world
+        PlayerList.LOGGER.info("{}[{}] logged in with entity id {} at ([{}]{}, {}, {})", serverplayer.getPlainTextName(), s1, serverplayer.getId(), serverlevel.serverLevelData.getLevelName(), serverplayer.getX(), serverplayer.getY(), serverplayer.getZ());
     }
 
     public void updateEntireScoreboard(ServerScoreboard serverscoreboard, ServerPlayer serverplayer) {
@@ -259,10 +352,10 @@
         });
     }
 
-    public Optional<CompoundTag> loadPlayerData(NameAndId nameandid) {
+    public Optional<CompoundTag> loadPlayerData(Player nameandid) { // CraftBukkit
         CompoundTag compoundtag = this.server.getWorldData().getLoadedPlayerTag();
 
-        if (this.server.isSingleplayerOwner(nameandid) && compoundtag != null) {
+        if (this.server.isSingleplayerOwner(nameandid.nameAndId()) && compoundtag != null) { // CraftBukkit
             PlayerList.LOGGER.debug("loading single player");
             return Optional.of(compoundtag);
         } else {
@@ -271,14 +364,15 @@
     }
 
     protected void save(ServerPlayer serverplayer) {
+        if (!serverplayer.getBukkitEntity().isPersistent()) return; // CraftBukkit
         this.playerIo.save(serverplayer);
-        ServerStatsCounter serverstatscounter = (ServerStatsCounter) this.stats.get(serverplayer.getUUID());
+        ServerStatsCounter serverstatscounter = (ServerStatsCounter) serverplayer.getStats(); // CraftBukkit
 
         if (serverstatscounter != null) {
             serverstatscounter.save();
         }
 
-        PlayerAdvancements playeradvancements = (PlayerAdvancements) this.advancements.get(serverplayer.getUUID());
+        PlayerAdvancements playeradvancements = (PlayerAdvancements) serverplayer.getAdvancements(); // CraftBukkit
 
         if (playeradvancements != null) {
             playeradvancements.save();
@@ -286,10 +380,24 @@
 
     }
 
-    public void remove(ServerPlayer serverplayer) {
+    public String remove(ServerPlayer serverplayer) { // CraftBukkit - return string
         ServerLevel serverlevel = serverplayer.level();
 
         serverplayer.awardStat(Stats.LEAVE_GAME);
+
+        // CraftBukkit start - Quitting must be before we do final save of data, in case plugins need to modify it
+        // See SPIGOT-5799, SPIGOT-6145
+        if (serverplayer.containerMenu != serverplayer.inventoryMenu) {
+            serverplayer.closeContainer();
+        }
+
+        PlayerQuitEvent playerQuitEvent = new PlayerQuitEvent(serverplayer.getBukkitEntity(), serverplayer.kickLeaveMessage != null ? serverplayer.kickLeaveMessage : "\u00A7e" + serverplayer.getScoreboardName() + " left the game");
+        cserver.getPluginManager().callEvent(playerQuitEvent);
+        serverplayer.getBukkitEntity().disconnect(playerQuitEvent.getQuitMessage());
+
+        serverplayer.doTick(); // SPIGOT-924
+        // CraftBukkit end
+
         this.save(serverplayer);
         if (serverplayer.isPassenger()) {
             Entity entity = serverplayer.getRootVehicle();
@@ -298,7 +406,7 @@
                 PlayerList.LOGGER.debug("Removing player mount");
                 serverplayer.stopRiding();
                 entity.getPassengersAndSelf().forEach((entity1) -> {
-                    entity1.setRemoved(Entity.RemovalReason.UNLOADED_WITH_PLAYER);
+                    entity1.setRemoved(Entity.RemovalReason.UNLOADED_WITH_PLAYER, EntityRemoveEvent.Cause.PLAYER_QUIT); // CraftBukkit - add Bukkit remove cause
                 });
             }
         }
@@ -306,7 +414,7 @@
         serverplayer.unRide();
 
         for (ThrownEnderpearl thrownenderpearl : serverplayer.getEnderPearls()) {
-            thrownenderpearl.setRemoved(Entity.RemovalReason.UNLOADED_WITH_PLAYER);
+            thrownenderpearl.setRemoved(Entity.RemovalReason.UNLOADED_WITH_PLAYER, EntityRemoveEvent.Cause.PLAYER_QUIT); // CraftBukkit - add Bukkit remove cause
         }
 
         serverlevel.removePlayerImmediately(serverplayer, Entity.RemovalReason.UNLOADED_WITH_PLAYER);
@@ -318,15 +426,60 @@
 
         if (serverplayer1 == serverplayer) {
             this.playersByUUID.remove(uuid);
-            this.stats.remove(uuid);
-            this.advancements.remove(uuid);
+            // CraftBukkit start
+            // this.stats.remove(uuid);
+            // this.advancements.remove(uuid);
+            // CraftBukkit end
             this.server.notificationManager().playerLeft(serverplayer);
         }
 
-        this.broadcastAll(new ClientboundPlayerInfoRemovePacket(List.of(serverplayer.getUUID())));
+        // CraftBukkit start
+        // this.broadcastAll(new ClientboundPlayerInfoRemovePacket(List.of(entityplayer.getUUID())));
+        ClientboundPlayerInfoRemovePacket packet = new ClientboundPlayerInfoRemovePacket(List.of(serverplayer.getUUID()));
+        for (int i = 0; i < players.size(); i++) {
+            ServerPlayer serverplayer2 = (ServerPlayer) this.players.get(i);
+
+            if (serverplayer2.getBukkitEntity().canSee(serverplayer.getBukkitEntity())) {
+                serverplayer2.connection.send(packet);
+            } else {
+                serverplayer2.getBukkitEntity().onEntityRemove(serverplayer);
+            }
+        }
+        // This removes the scoreboard (and player reference) for the specific player in the manager
+        cserver.getScoreboardManager().removePlayer(serverplayer.getBukkitEntity());
+        // CraftBukkit end
+
+        return playerQuitEvent.getQuitMessage(); // CraftBukkit
     }
 
-    public @Nullable Component canPlayerLogin(SocketAddress socketaddress, NameAndId nameandid) {
+    // CraftBukkit start - Whole method, SocketAddress to LoginListener, added hostname to signature, return EntityPlayer
+    public @Nullable ServerPlayer canPlayerLogin(ServerLoginPacketListenerImpl serverloginpacketlistenerimpl, GameProfile gameprofile) {
+        // Moved from processLogin
+        NameAndId nameandid = new NameAndId(gameprofile);
+        UUID uuid = gameprofile.id();
+        Set<ServerPlayer> set = Sets.newIdentityHashSet();
+
+        for (ServerPlayer serverplayer : this.players) {
+            if (serverplayer.getUUID().equals(uuid)) {
+                set.add(serverplayer);
+            }
+        }
+
+        for (ServerPlayer serverplayer2 : set) {
+            save(serverplayer2); // CraftBukkit - Force the player's inventory to be saved
+            serverplayer2.connection.disconnect(PlayerList.DUPLICATE_LOGIN_DISCONNECT_MESSAGE);
+        }
+
+        // Instead of kicking then returning, we need to store the kick reason
+        // in the event, check with plugins to see if it's ok, and THEN kick
+        // depending on the outcome.
+        SocketAddress socketaddress = serverloginpacketlistenerimpl.connection.getRemoteAddress();
+
+        ServerPlayer entity = new ServerPlayer(this.server, this.server.getLevel(Level.OVERWORLD), gameprofile, ClientInformation.createDefault());
+        entity.transferCookieConnection = serverloginpacketlistenerimpl;
+        org.bukkit.entity.Player player = entity.getBukkitEntity();
+        PlayerLoginEvent event = new PlayerLoginEvent(player, serverloginpacketlistenerimpl.connection.hostname, ((java.net.InetSocketAddress) socketaddress).getAddress());
+
         if (this.bans.isBanned(nameandid)) {
             UserBanListEntry userbanlistentry = (UserBanListEntry) this.bans.get(nameandid);
             MutableComponent mutablecomponent = Component.translatable("multiplayer.disconnect.banned.reason", userbanlistentry.getReasonMessage());
@@ -335,9 +488,11 @@
                 mutablecomponent.append((Component) Component.translatable("multiplayer.disconnect.banned.expiration", PlayerList.BAN_DATE_FORMAT.format(userbanlistentry.getExpires())));
             }
 
-            return mutablecomponent;
+            // return ichatmutablecomponent;
+            event.disallow(PlayerLoginEvent.Result.KICK_BANNED, CraftChatMessage.fromComponent(mutablecomponent));
         } else if (!this.isWhiteListed(nameandid)) {
-            return Component.translatable("multiplayer.disconnect.not_whitelisted");
+            MutableComponent mutablecomponent = Component.translatable("multiplayer.disconnect.not_whitelisted");
+            event.disallow(PlayerLoginEvent.Result.KICK_WHITELIST, CraftChatMessage.fromComponent(mutablecomponent));
         } else if (this.ipBans.isBanned(socketaddress)) {
             IpBanListEntry ipbanlistentry = this.ipBans.get(socketaddress);
             MutableComponent mutablecomponent1 = Component.translatable("multiplayer.disconnect.banned_ip.reason", ipbanlistentry.getReasonMessage());
@@ -346,13 +501,25 @@
                 mutablecomponent1.append((Component) Component.translatable("multiplayer.disconnect.banned_ip.expiration", PlayerList.BAN_DATE_FORMAT.format(ipbanlistentry.getExpires())));
             }
 
-            return mutablecomponent1;
+            // return ichatmutablecomponent1;
+            event.disallow(PlayerLoginEvent.Result.KICK_BANNED, CraftChatMessage.fromComponent(mutablecomponent1));
         } else {
-            return this.players.size() >= this.getMaxPlayers() && !this.canBypassPlayerLimit(nameandid) ? Component.translatable("multiplayer.disconnect.server_full") : null;
+            // return this.players.size() >= this.getMaxPlayers() && !this.canBypassPlayerLimit(nameandid) ? IChatBaseComponent.translatable("multiplayer.disconnect.server_full") : null;
+            if (this.players.size() >= this.getMaxPlayers() && !this.canBypassPlayerLimit(nameandid)) {
+                event.disallow(PlayerLoginEvent.Result.KICK_FULL, "The server is full");
+            }
         }
+
+        cserver.getPluginManager().callEvent(event);
+        if (event.getResult() != PlayerLoginEvent.Result.ALLOWED) {
+            serverloginpacketlistenerimpl.disconnect(event.getKickMessage());
+            return null;
+        }
+        return entity;
     }
 
-    public boolean disconnectAllPlayersWithProfile(UUID uuid) {
+    public boolean disconnectAllPlayersWithProfile(UUID uuid, ServerPlayer player) { // CraftBukkit - added EntityPlayer
+        /* CraftBukkit startMoved up
         Set<ServerPlayer> set = Sets.newIdentityHashSet();
 
         for (ServerPlayer serverplayer : this.players) {
@@ -372,33 +539,67 @@
         }
 
         return !set.isEmpty();
+        */
+        return player == null;
+        // CraftBukkit end
     }
 
-    public ServerPlayer respawn(ServerPlayer serverplayer, boolean flag, Entity.RemovalReason entity_removalreason) {
-        TeleportTransition teleporttransition = serverplayer.findRespawnPositionAndUseSpawnBlock(!flag, TeleportTransition.DO_NOTHING);
+    // CraftBukkit start
+    public ServerPlayer respawn(ServerPlayer serverplayer, boolean flag, Entity.RemovalReason entity_removalreason, RespawnReason reason) {
+        return this.respawn(serverplayer, flag, entity_removalreason, reason, null);
+    }
 
+    public ServerPlayer respawn(ServerPlayer serverplayer, boolean flag, Entity.RemovalReason entity_removalreason, RespawnReason reason, Location location) {
+        // TeleportTransition teleporttransition = serverplayer.findRespawnPositionAndUseSpawnBlock(!flag, TeleportTransition.DO_NOTHING); // CraftBukkit
+
+        serverplayer.stopRiding(); // CraftBukkit
         this.players.remove(serverplayer);
         serverplayer.level().removePlayerImmediately(serverplayer, entity_removalreason);
+        /* CraftBukkit start
         ServerLevel serverlevel = teleporttransition.newLevel();
         ServerPlayer serverplayer1 = new ServerPlayer(this.server, serverlevel, serverplayer.getGameProfile(), serverplayer.clientInformation());
+        // */
+        ServerPlayer serverplayer1 = serverplayer;
+        Level fromWorld = serverplayer.level();
+        serverplayer.wonGame = false;
+        // CraftBukkit end
 
         serverplayer1.connection = serverplayer.connection;
         serverplayer1.restoreFrom(serverplayer, flag);
         serverplayer1.setId(serverplayer.getId());
         serverplayer1.setMainArm(serverplayer.getMainArm());
+        // CraftBukkit - not required, just copies old location into reused entity
+        /*
         if (!teleporttransition.missingRespawnBlock()) {
             serverplayer1.copyRespawnPosition(serverplayer);
         }
+         */
+        // CraftBukkit end
 
         for (String s : serverplayer.getTags()) {
             serverplayer1.addTag(s);
         }
 
+        // CraftBukkit start - fire PlayerRespawnEvent
+        TeleportTransition teleporttransition;
+        if (location == null) {
+            teleporttransition = serverplayer.findRespawnPositionAndUseSpawnBlock(!flag, TeleportTransition.DO_NOTHING, reason);
+
+            if (!flag) serverplayer.reset(); // SPIGOT-4785
+        } else {
+            teleporttransition = new TeleportTransition(((CraftWorld) location.getWorld()).getHandle(), CraftLocation.toVec3D(location), Vec3.ZERO, location.getYaw(), location.getPitch(), TeleportTransition.DO_NOTHING);
+        }
+        ServerLevel serverlevel = teleporttransition.newLevel();
+        serverplayer1.spawnIn(serverlevel, flag);
+        serverplayer1.unsetRemoved();
+        serverplayer1.setShiftKeyDown(false);
         Vec3 vec3 = teleporttransition.position();
 
-        serverplayer1.snapTo(vec3.x, vec3.y, vec3.z, teleporttransition.yRot(), teleporttransition.xRot());
+        serverplayer1.forceSetPositionRotation(vec3.x, vec3.y, vec3.z, teleporttransition.yRot(), teleporttransition.xRot());
+        // CraftBukkit end
         if (teleporttransition.missingRespawnBlock()) {
             serverplayer1.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.NO_RESPAWN_BLOCK_AVAILABLE, 0.0F));
+            serverplayer1.setRespawnPosition(null, false, PlayerSpawnChangeEvent.Cause.RESET); // CraftBukkit - SPIGOT-5988: Clear respawn location when obstructed
         }
 
         byte b0 = (byte) (flag ? 1 : 0);
@@ -406,17 +607,19 @@
         LevelData leveldata = serverlevel1.getLevelData();
 
         serverplayer1.connection.send(new ClientboundRespawnPacket(serverplayer1.createCommonSpawnInfo(serverlevel1), b0));
-        serverplayer1.connection.teleport(serverplayer1.getX(), serverplayer1.getY(), serverplayer1.getZ(), serverplayer1.getYRot(), serverplayer1.getXRot());
+        serverplayer1.connection.teleport(CraftLocation.toBukkit(serverplayer1.position(), serverlevel1.getWorld(), serverplayer1.getYRot(), serverplayer1.getXRot())); // CraftBukkit
         serverplayer1.connection.send(new ClientboundSetDefaultSpawnPositionPacket(serverlevel.getRespawnData()));
         serverplayer1.connection.send(new ClientboundChangeDifficultyPacket(leveldata.getDifficulty(), leveldata.isDifficultyLocked()));
         serverplayer1.connection.send(new ClientboundSetExperiencePacket(serverplayer1.experienceProgress, serverplayer1.totalExperience, serverplayer1.experienceLevel));
         this.sendActivePlayerEffects(serverplayer1);
         this.sendLevelInfo(serverplayer1, serverlevel);
         this.sendPlayerPermissionLevel(serverplayer1);
-        serverlevel.addRespawnedPlayer(serverplayer1);
-        this.players.add(serverplayer1);
-        this.playersByUUID.put(serverplayer1.getUUID(), serverplayer1);
-        serverplayer1.initInventoryMenu();
+        if (!serverplayer.connection.isDisconnected()) {
+            serverlevel.addRespawnedPlayer(serverplayer1);
+            this.players.add(serverplayer1);
+            this.playersByUUID.put(serverplayer1.getUUID(), serverplayer1);
+        }
+        // serverplayer1.initInventoryMenu();
         serverplayer1.setHealth(serverplayer1.getHealth());
         ServerPlayer.RespawnConfig serverplayer_respawnconfig = serverplayer1.getRespawnConfig();
 
@@ -433,6 +636,27 @@
                 }
             }
         }
+        // Added from changeDimension
+        sendAllPlayerInfo(serverplayer); // Update health, etc...
+        serverplayer.onUpdateAbilities();
+        for (MobEffectInstance mobEffect : serverplayer.getActiveEffects()) {
+            serverplayer.connection.send(new ClientboundUpdateMobEffectPacket(serverplayer.getId(), mobEffect, false)); // blend = false
+        }
+
+        // Fire advancement trigger
+        serverplayer.triggerDimensionChangeTriggers(serverlevel);
+
+        // Don't fire on respawn
+        if (fromWorld != serverlevel) {
+            PlayerChangedWorldEvent event = new PlayerChangedWorldEvent(serverplayer.getBukkitEntity(), fromWorld.getWorld());
+            server.server.getPluginManager().callEvent(event);
+        }
+
+        // Save player file again if they were disconnected
+        if (serverplayer.connection.isDisconnected()) {
+            this.save(serverplayer);
+        }
+        // CraftBukkit end
 
         return serverplayer1;
     }
@@ -456,7 +680,18 @@
 
     public void tick() {
         if (++this.sendAllPlayerInfoIn > 600) {
-            this.broadcastAll(new ClientboundPlayerInfoUpdatePacket(EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_LATENCY), this.players));
+            // CraftBukkit start
+            for (int i = 0; i < this.players.size(); ++i) {
+                final ServerPlayer target = (ServerPlayer) this.players.get(i);
+
+                target.connection.send(new ClientboundPlayerInfoUpdatePacket(EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_LATENCY), this.players.stream().filter(new Predicate<ServerPlayer>() {
+                    @Override
+                    public boolean test(ServerPlayer input) {
+                        return target.getBukkitEntity().canSee(input.getBukkitEntity());
+                    }
+                }).collect(Collectors.toList())));
+            }
+            // CraftBukkit end
             this.sendAllPlayerInfoIn = 0;
         }
 
@@ -469,6 +704,25 @@
 
     }
 
+    // CraftBukkit start - add a world/entity limited version
+    public void broadcastAll(Packet packet, Player player) {
+        for (int i = 0; i < this.players.size(); ++i) {
+            ServerPlayer serverplayer =  this.players.get(i);
+            if (player != null && !serverplayer.getBukkitEntity().canSee(player.getBukkitEntity())) {
+                continue;
+            }
+            ((ServerPlayer) this.players.get(i)).connection.send(packet);
+        }
+    }
+
+    public void broadcastAll(Packet packet, Level level) {
+        for (int i = 0; i < level.players().size(); ++i) {
+            ((ServerPlayer) level.players().get(i)).connection.send(packet);
+        }
+
+    }
+    // CraftBukkit end
+
     public void broadcastAll(Packet<?> packet, ResourceKey<Level> resourcekey) {
         for (ServerPlayer serverplayer : this.players) {
             if (serverplayer.level().dimension() == resourcekey) {
@@ -582,6 +836,7 @@
             serverplayer.connection.send(new ClientboundEntityEventPacket(serverplayer, b1));
         }
 
+        serverplayer.getBukkitEntity().recalculatePermissions(); // CraftBukkit
         this.server.getCommands().sendCommands(serverplayer);
     }
 
@@ -611,6 +866,12 @@
         for (int i = 0; i < this.players.size(); ++i) {
             ServerPlayer serverplayer = (ServerPlayer) this.players.get(i);
 
+            // CraftBukkit start - Test if player receiving packet can see the source of the packet
+            if (player != null && !serverplayer.getBukkitEntity().canSee(player.getBukkitEntity())) {
+               continue;
+            }
+            // CraftBukkit end
+
             if (serverplayer != player && serverplayer.level().dimension() == resourcekey) {
                 double d4 = d0 - serverplayer.getX();
                 double d5 = d1 - serverplayer.getY();
@@ -656,9 +917,13 @@
         serverplayer.connection.send(new ClientboundSetTimePacket(serverlevel.getGameTime(), serverlevel.getDayTime(), (Boolean) serverlevel.getGameRules().get(GameRules.ADVANCE_TIME)));
         serverplayer.connection.send(new ClientboundSetDefaultSpawnPositionPacket(serverlevel.getRespawnData()));
         if (serverlevel.isRaining()) {
-            serverplayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.START_RAINING, 0.0F));
-            serverplayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.RAIN_LEVEL_CHANGE, serverlevel.getRainLevel(1.0F)));
-            serverplayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.THUNDER_LEVEL_CHANGE, serverlevel.getThunderLevel(1.0F)));
+            // CraftBukkit start - handle player weather
+            // entityplayer.connection.send(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.START_RAINING, 0.0F));
+            // entityplayer.connection.send(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.RAIN_LEVEL_CHANGE, worldserver.getRainLevel(1.0F)));
+            // entityplayer.connection.send(new PacketPlayOutGameStateChange(PacketPlayOutGameStateChange.THUNDER_LEVEL_CHANGE, worldserver.getThunderLevel(1.0F)));
+            serverplayer.setPlayerWeather(org.bukkit.WeatherType.DOWNFALL, false);
+            serverplayer.updateWeather(-serverlevel.rainLevel, serverlevel.rainLevel, -serverlevel.thunderLevel, serverlevel.thunderLevel);
+            // CraftBukkit end
         }
 
         serverplayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.LEVEL_CHUNKS_LOAD_START, 0.0F));
@@ -667,8 +932,16 @@
 
     public void sendAllPlayerInfo(ServerPlayer serverplayer) {
         serverplayer.inventoryMenu.sendAllDataToRemote();
-        serverplayer.resetSentInfo();
+        // entityplayer.resetSentInfo();
+        serverplayer.getBukkitEntity().updateScaledHealth(); // CraftBukkit - Update scaled health on respawn and worldchange
+        serverplayer.refreshEntityData(serverplayer); // CraftBukkkit - SPIGOT-7218: sync metadata
         serverplayer.connection.send(new ClientboundSetHeldSlotPacket(serverplayer.getInventory().getSelectedSlot()));
+        // CraftBukkit start - from GameRules
+        int i = serverplayer.level().getGameRules().get(GameRules.REDUCED_DEBUG_INFO) ? 22 : 23;
+        serverplayer.connection.send(new ClientboundEntityEventPacket(serverplayer, (byte) i));
+        float immediateRespawn = serverplayer.level().getGameRules().get(GameRules.IMMEDIATE_RESPAWN) ? 1.0F: 0.0F;
+        serverplayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.IMMEDIATE_RESPAWN, immediateRespawn));
+        // CraftBukkit end
     }
 
     public int getPlayerCount() {
@@ -716,12 +989,22 @@
     }
 
     public void removeAll() {
-        for (int i = 0; i < this.players.size(); ++i) {
-            ((ServerPlayer) this.players.get(i)).connection.disconnect((Component) Component.translatable("multiplayer.disconnect.server_shutdown"));
+        // CraftBukkit start - disconnect safely
+        for (ServerPlayer player : this.players) {
+            player.connection.disconnect(CraftChatMessage.fromStringOrEmpty(this.server.server.getShutdownMessage())); // CraftBukkit - add custom shutdown message
         }
+        // CraftBukkit end
 
     }
 
+    // CraftBukkit start
+    public void broadcastMessage(Component[] iChatBaseComponents) {
+        for (Component component : iChatBaseComponents) {
+            broadcastSystemMessage(component, false);
+        }
+    }
+    // CraftBukkit end
+
     public void broadcastSystemMessage(Component component, boolean flag) {
         this.broadcastSystemMessage(component, (serverplayer) -> {
             return component;
@@ -776,16 +1059,29 @@
     }
 
     public ServerStatsCounter getPlayerStats(Player player) {
-        GameProfile gameprofile = player.getGameProfile();
+        // CraftBukkit start
+        return getPlayerStats(player.nameAndId());
+    }
 
-        return (ServerStatsCounter) this.stats.computeIfAbsent(gameprofile.id(), (uuid) -> {
+    public ServerStatsCounter getPlayerStats(ServerPlayer entityhuman) {
+        NameAndId gameprofile = entityhuman.nameAndId();
+        ServerStatsCounter serverstatscounter = entityhuman.getStats();
+        return serverstatscounter == null ? getPlayerStats(gameprofile) : serverstatscounter;
+    }
+
+    public ServerStatsCounter getPlayerStats(NameAndId gameprofile) {
+        ServerPlayer entityhuman = this.getPlayer(gameprofile.id());
+        ServerStatsCounter serverstatscounter = entityhuman == null ? null : (ServerStatsCounter) entityhuman.getStats();
+        if (serverstatscounter == null) {
             Path path = this.locateStatsFile(gameprofile);
 
             return new ServerStatsCounter(this.server, path);
-        });
+        }
+        return serverstatscounter;
     }
 
-    private Path locateStatsFile(GameProfile gameprofile) {
+    private Path locateStatsFile(NameAndId gameprofile) {
+        // CraftBukkit end
         Path path = this.server.getWorldPath(LevelResource.PLAYER_STATS_DIR);
         Path path1 = path.resolve(String.valueOf(gameprofile.id()) + ".json");
 
@@ -813,13 +1109,13 @@
 
     public PlayerAdvancements getPlayerAdvancements(ServerPlayer serverplayer) {
         UUID uuid = serverplayer.getUUID();
-        PlayerAdvancements playeradvancements = (PlayerAdvancements) this.advancements.get(uuid);
+        PlayerAdvancements playeradvancements = (PlayerAdvancements) serverplayer.getAdvancements(); // CraftBukkit
 
         if (playeradvancements == null) {
             Path path = this.server.getWorldPath(LevelResource.PLAYER_ADVANCEMENTS_DIR).resolve(String.valueOf(uuid) + ".json");
 
             playeradvancements = new PlayerAdvancements(this.server.getFixerUpper(), this, this.server.getAdvancements(), path, serverplayer);
-            this.advancements.put(uuid, playeradvancements);
+            // this.advancements.put(uuid, advancementdataplayer); // CraftBukkit
         }
 
         playeradvancements.setPlayer(serverplayer);
@@ -869,11 +1165,24 @@
     }
 
     public void reloadResources() {
-        for (PlayerAdvancements playeradvancements : this.advancements.values()) {
+        // CraftBukkit start
+        /*for (PlayerAdvancements playeradvancements : this.advancements.values()) {
             playeradvancements.reload(this.server.getAdvancements());
+        }*/
+
+        for (ServerPlayer player : players) {
+            player.getAdvancements().reload(this.server.getAdvancements());
+            player.getAdvancements().flushDirty(player, false); // CraftBukkit - trigger immediate flush of advancements
         }
+        // CraftBukkit end
 
         this.broadcastAll(new ClientboundUpdateTagsPacket(TagNetworkSerialization.serializeTagsToNetwork(this.registries)));
+        // CraftBukkit start
+        reloadRecipes();
+    }
+
+    public void reloadRecipes() {
+        // CraftBukkit end
         RecipeManager recipemanager = this.server.getRecipeManager();
         ClientboundUpdateRecipesPacket clientboundupdaterecipespacket = new ClientboundUpdateRecipesPacket(recipemanager.getSynchronizedItemProperties(), recipemanager.getSynchronizedStonecutterRecipes());
 
