--- a/net/minecraft/client/player/RemotePlayer.java
+++ b/net/minecraft/client/player/RemotePlayer.java
@@ -44,6 +_,10 @@
 
     @Override
     public void aiStep() {
+        if (this.isInterpolating()) {
+            this.getInterpolation().interpolate();
+        }
+
         if (this.lerpHeadSteps > 0) {
             this.lerpHeadRotationStep(this.lerpHeadSteps, this.lerpYHeadRot);
             this.lerpHeadSteps--;
@@ -60,6 +_,7 @@
             this.lerpDeltaMovementSteps--;
         }
 
+        this.updateSwingTime();
         this.updateBob();
 
         try (Zone ignored = Profiler.get().zone("push")) {
@@ -70,11 +_,7 @@
     @Override
     public void lerpMotion(Vec3 movement) {
         this.lerpDeltaMovement = movement;
-        if (this.getType().hasUpdateInterval()) {
-            this.lerpDeltaMovementSteps = this.getType().updateInterval() + 1;
-        } else {
-            this.lerpDeltaMovementSteps = 1;
-        }
+        this.lerpDeltaMovementSteps = this.getType().updateInterval() + 1;
     }
 
     @Override
