--- a/net/minecraft/client/resources/model/cuboid/CuboidModelElement.java
+++ b/net/minecraft/client/resources/model/cuboid/CuboidModelElement.java
@@ -31,7 +_,7 @@
         this(from, to, faces, null, true, 0);
     }
 
-    public static class Deserializer implements JsonDeserializer<CuboidModelElement> {
+    protected static class Deserializer implements JsonDeserializer<CuboidModelElement> {
         private static final boolean DEFAULT_SHADE = true;
         private static final int DEFAULT_LIGHT_EMISSION = 0;
         private static final String FIELD_SHADE = "shade";
