--- a/net/minecraft/client/resources/model/cuboid/CuboidFace.java
+++ b/net/minecraft/client/resources/model/cuboid/CuboidFace.java
@@ -26,7 +_,7 @@
         return uvs.getVertexV(rotation.rotateVertexIndex(index)) / 16.0F;
     }
 
-    public static class Deserializer implements JsonDeserializer<CuboidFace> {
+    protected static class Deserializer implements JsonDeserializer<CuboidFace> {
         private static final int DEFAULT_TINT_INDEX = -1;
         private static final int DEFAULT_ROTATION = 0;
 
