--- a/net/minecraft/client/resources/model/cuboid/FaceBakery.java
+++ b/net/minecraft/client/resources/model/cuboid/FaceBakery.java
@@ -26,7 +_,7 @@
     private static final Vector3fc BLOCK_MIDDLE = new Vector3f(0.5F, 0.5F, 0.5F);
 
     @VisibleForTesting
-    public static CuboidFace.UVs defaultFaceUV(Vector3fc from, Vector3fc to, Direction facing) {
+    static CuboidFace.UVs defaultFaceUV(Vector3fc from, Vector3fc to, Direction facing) {
         return switch (facing) {
             case DOWN -> new CuboidFace.UVs(from.x(), 16.0F - to.z(), to.x(), 16.0F - from.z());
             case UP -> new CuboidFace.UVs(from.x(), from.z(), to.x(), to.z());
@@ -207,7 +_,7 @@
         return best;
     }
 
-    public static void recalculateWinding(Vector3fc[] positions, long[] uvs, Direction direction) {
+    private static void recalculateWinding(Vector3fc[] positions, long[] uvs, Direction direction) {
         float minX = 999.0F;
         float minY = 999.0F;
         float minZ = 999.0F;
