--- a/net/minecraft/client/resources/model/cuboid/ItemTransform.java
+++ b/net/minecraft/client/resources/model/cuboid/ItemTransform.java
@@ -47,10 +_,10 @@
         }
     }
 
-    public static class Deserializer implements JsonDeserializer<ItemTransform> {
-        public static final Vector3fc DEFAULT_ROTATION = new Vector3f();
-        public static final Vector3fc DEFAULT_TRANSLATION = new Vector3f();
-        public static final Vector3fc DEFAULT_SCALE = new Vector3f(1.0F, 1.0F, 1.0F);
+    protected static class Deserializer implements JsonDeserializer<ItemTransform> {
+        private static final Vector3fc DEFAULT_ROTATION = new Vector3f();
+        private static final Vector3fc DEFAULT_TRANSLATION = new Vector3f();
+        private static final Vector3fc DEFAULT_SCALE = new Vector3f(1.0F, 1.0F, 1.0F);
         public static final float MAX_TRANSLATION = 5.0F;
         public static final float MAX_SCALE = 4.0F;
 
