--- a/net/minecraft/client/resources/model/cuboid/ItemModelGenerator.java
+++ b/net/minecraft/client/resources/model/cuboid/ItemModelGenerator.java
@@ -90,7 +_,7 @@
         bakeSideFaces(builder, interner, modelState, materialInfo);
     }
 
-    public static void bakeSideFaces(QuadCollection.Builder builder, ModelBaker.Interner interner, ModelState modelState, BakedQuad.MaterialInfo materialInfo) {
+    private static void bakeSideFaces(QuadCollection.Builder builder, ModelBaker.Interner interner, ModelState modelState, BakedQuad.MaterialInfo materialInfo) {
         SpriteContents sprite = materialInfo.sprite().contents();
         float xScale = 16.0F / sprite.width();
         float yScale = 16.0F / sprite.height();
@@ -206,7 +_,7 @@
         return x >= 0 && y >= 0 && x < width && y < height ? sprite.isTransparent(frame, x, y) : true;
     }
 
-    public record ItemLayerKey(Material.Baked material, ModelState modelState, int layerIndex) implements ModelBaker.SharedOperationKey<QuadCollection> {
+    private record ItemLayerKey(Material.Baked material, ModelState modelState, int layerIndex) implements ModelBaker.SharedOperationKey<QuadCollection> {
         public QuadCollection compute(ModelBaker modelBakery) {
             QuadCollection.Builder builder = new QuadCollection.Builder();
             BakedQuad.MaterialInfo materialInfo = modelBakery.interner()
