--- a/net/minecraft/client/resources/model/cuboid/ItemTransforms.java
+++ b/net/minecraft/client/resources/model/cuboid/ItemTransforms.java
@@ -49,7 +_,7 @@
         };
     }
 
-    public static class Deserializer implements JsonDeserializer<ItemTransforms> {
+    protected static class Deserializer implements JsonDeserializer<ItemTransforms> {
         public ItemTransforms deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
             JsonObject object = json.getAsJsonObject();
             ItemTransform thirdPersonRightHand = this.getTransform(context, object, ItemDisplayContext.THIRD_PERSON_RIGHT_HAND);
