--- a/net/minecraft/client/renderer/blockentity/BeaconRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/BeaconRenderer.java
@@ -93,7 +_,7 @@
         float texVOff = Mth.frac(scroll * 0.2F - Mth.floor(scroll * 0.1F));
         poseStack.pushPose();
         {
-        poseStack.rotateDegrees(Axis.YP, animationTime * 2.25F - 45.0F);
+        poseStack.mulPose(Axis.YP.rotationDegrees(animationTime * 2.25F - 45.0F));
         float wnx = 0.0F;
         float wnz = solidBeamRadius;
         float enx = solidBeamRadius;
