--- a/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
+++ b/net/minecraft/client/renderer/blockentity/BlockEntityRenderers.java
@@ -16,7 +_,7 @@
 public class BlockEntityRenderers {
     private static final Map<BlockEntityType<?>, BlockEntityRendererProvider<?, ?>> PROVIDERS = Maps.newHashMap();
 
-    public static <T extends BlockEntity, S extends BlockEntityRenderState> void register(
+    private static <T extends BlockEntity, S extends BlockEntityRenderState> void register(
         BlockEntityType<? extends T> type, BlockEntityRendererProvider<T, S> renderer
     ) {
         PROVIDERS.put(type, renderer);
