--- a/net/minecraft/client/renderer/blockentity/BlockEntityRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/BlockEntityRenderer.java
@@ -11,10 +_,8 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.extensions.IBlockEntityRendererExtension;
-
 @OnlyIn(Dist.CLIENT)
-public interface BlockEntityRenderer<T extends BlockEntity, S extends BlockEntityRenderState> extends IBlockEntityRendererExtension<T> {
+public interface BlockEntityRenderer<T extends BlockEntity, S extends BlockEntityRenderState> {
     S createRenderState();
 
     default void extractRenderState(
