--- a/net/minecraft/client/renderer/blockentity/BrushableBlockRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/BrushableBlockRenderer.java
@@ -58,9 +_,9 @@
             poseStack.translate(0.0F, 0.5F, 0.0F);
             float[] translations = this.translations(state.hitDirection, state.dustProgress);
             poseStack.translate(translations[0], translations[1], translations[2]);
-            poseStack.rotateDegrees(Axis.YP, 75.0F);
+            poseStack.mulPose(Axis.YP.rotationDegrees(75.0F));
             boolean eastWest = state.hitDirection == Direction.EAST || state.hitDirection == Direction.WEST;
-            poseStack.rotateDegrees(Axis.YP, (eastWest ? 90 : 0) + 11);
+            poseStack.mulPose(Axis.YP.rotationDegrees((eastWest ? 90 : 0) + 11));
             poseStack.scale(0.5F, 0.5F, 0.5F);
             state.itemState.submit(poseStack, submitNodeCollector, state.lightCoords, OverlayTexture.NO_OVERLAY, 0);
             poseStack.popPose();
