--- a/net/minecraft/client/renderer/blockentity/LecternRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/LecternRenderer.java
@@ -48,8 +_,8 @@
         if (state.hasBook) {
             poseStack.pushPose();
             poseStack.translate(0.5F, 1.0625F, 0.5F);
-            poseStack.rotateDegrees(Axis.YP, -state.yRot);
-            poseStack.rotateDegrees(Axis.ZP, 67.5F);
+            poseStack.mulPose(Axis.YP.rotationDegrees(-state.yRot));
+            poseStack.mulPose(Axis.ZP.rotationDegrees(67.5F));
             poseStack.translate(0.0F, -0.125F, 0.0F);
             submitNodeCollector.submitModel(
                 this.bookModel, BOOK_STATE, poseStack, state.lightCoords, OverlayTexture.NO_OVERLAY, -1, EnchantTableRenderer.BOOK_TEXTURE, this.sprites, 0
