--- a/net/minecraft/client/renderer/blockentity/SpawnerRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/SpawnerRenderer.java
@@ -60,9 +_,9 @@
     ) {
         poseStack.pushPose();
         poseStack.translate(0.5F, 0.4F, 0.5F);
-        poseStack.rotateDegrees(Axis.YP, spin);
+        poseStack.mulPose(Axis.YP.rotationDegrees(spin));
         poseStack.translate(0.0F, -0.2F, 0.0F);
-        poseStack.rotateDegrees(Axis.XP, -30.0F);
+        poseStack.mulPose(Axis.XP.rotationDegrees(-30.0F));
         poseStack.scale(scale, scale, scale);
         entityRenderer.submit(displayEntity, camera, 0.0, 0.0, 0.0, poseStack, submitNodeCollector);
         poseStack.popPose();
