--- a/net/minecraft/client/renderer/blockentity/VaultRenderer.java
+++ b/net/minecraft/client/renderer/blockentity/VaultRenderer.java
@@ -56,7 +_,7 @@
         if (state.displayItem != null) {
             poseStack.pushPose();
             poseStack.translate(0.5F, 0.4F, 0.5F);
-            poseStack.rotateDegrees(Axis.YP, state.spin);
+            poseStack.mulPose(Axis.YP.rotationDegrees(state.spin));
             ItemEntityRenderer.renderMultipleFromCount(poseStack, submitNodeCollector, state.lightCoords, state.displayItem, this.random);
             poseStack.popPose();
         }
