--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -50,6 +50,10 @@
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityDamageEvent;
+// CraftBukkit end
+
 public class Armadillo extends Animal {
 
     public static final float BABY_SCALE = 0.6F;
@@ -130,16 +134,18 @@
         ProfilerFiller profilerfiller = Profiler.get();
 
         profilerfiller.push("armadilloBrain");
-        this.brain.tick(serverlevel, this);
+        ((Brain<Armadillo>) this.brain).tick(serverlevel, this); // CraftBukkit - decompile error
         profilerfiller.pop();
         profilerfiller.push("armadilloActivityUpdate");
         ArmadilloAi.updateActivity(this);
         profilerfiller.pop();
         if (this.isAlive() && --this.scuteTime <= 0 && this.shouldDropLoot(serverlevel)) {
+            this.forceDrops = true; // CraftBukkit
             if (this.dropFromGiftLootTable(serverlevel, BuiltInLootTables.ARMADILLO_SHED, this::spawnAtLocation)) {
                 this.playSound(SoundEvents.ARMADILLO_SCUTE_DROP, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
                 this.gameEvent(GameEvent.ENTITY_PLACE);
             }
+            this.forceDrops = false; // CraftBukkit
 
             this.scuteTime = this.pickNextScuteDropTime();
         }
@@ -285,8 +291,13 @@
     }
 
     @Override
-    protected void actuallyHurt(ServerLevel serverlevel, DamageSource damagesource, float f) {
-        super.actuallyHurt(serverlevel, damagesource, f);
+    // CraftBukkit start - void -> boolean
+    public boolean actuallyHurt(ServerLevel serverlevel, DamageSource damagesource, float f, EntityDamageEvent event) {
+        boolean damageResult = super.actuallyHurt(serverlevel, damagesource, f, event);
+        if (!damageResult) {
+            return false;
+        }
+        // CraftBukkit end
         if (!this.isNoAi() && !this.isDeadOrDying()) {
             if (damagesource.getEntity() instanceof LivingEntity) {
                 this.getBrain().setMemoryWithExpiry(MemoryModuleType.DANGER_DETECTED_RECENTLY, true, 80L);
@@ -298,6 +309,7 @@
             }
 
         }
+        return true; // CraftBukkit
     }
 
     @Override
@@ -321,7 +333,9 @@
             if (level instanceof ServerLevel) {
                 ServerLevel serverlevel = (ServerLevel) level;
 
+                this.forceDrops = true; // CraftBukkit
                 this.dropFromEntityInteractLootTable(serverlevel, BuiltInLootTables.ARMADILLO_BRUSH, entity, itemstack, this::spawnAtLocation);
+                this.forceDrops = false; // CraftBukkit
                 this.playSound(SoundEvents.ARMADILLO_BRUSH);
                 this.gameEvent(GameEvent.ENTITY_INTERACT);
             }
