--- a/net/minecraft/data/worldgen/features/MiscOverworldFeatures.java
+++ b/net/minecraft/data/worldgen/features/MiscOverworldFeatures.java
@@ -23,11 +_,11 @@
 import net.minecraft.world.level.levelgen.feature.BlockBlobFeature;
 import net.minecraft.world.level.levelgen.feature.BlueIceFeature;
 import net.minecraft.world.level.levelgen.feature.BonusChestFeature;
+import net.minecraft.world.level.levelgen.feature.DesertWellFeature;
 import net.minecraft.world.level.levelgen.feature.DiskFeature;
 import net.minecraft.world.level.levelgen.feature.Feature;
 import net.minecraft.world.level.levelgen.feature.IcebergFeature;
 import net.minecraft.world.level.levelgen.feature.LakeFeature;
-import net.minecraft.world.level.levelgen.feature.OverlayFeature;
 import net.minecraft.world.level.levelgen.feature.SequenceFeature;
 import net.minecraft.world.level.levelgen.feature.SimpleBlockFeature;
 import net.minecraft.world.level.levelgen.feature.SnowAndFreezeFeature;
@@ -43,16 +_,7 @@
 import net.minecraft.world.level.levelgen.placement.EnvironmentScanPlacement;
 import net.minecraft.world.level.levelgen.placement.OffsetPlacement;
 import net.minecraft.world.level.levelgen.placement.PlacedFeature;
-import net.minecraft.world.level.levelgen.placement.RandomlySelectedPlacement;
-import net.minecraft.world.level.levelgen.structure.templatesystem.AlwaysTrueTest;
-import net.minecraft.world.level.levelgen.structure.templatesystem.BlockMatchTest;
-import net.minecraft.world.level.levelgen.structure.templatesystem.PosAlwaysTrueTest;
-import net.minecraft.world.level.levelgen.structure.templatesystem.ProcessorRule;
-import net.minecraft.world.level.levelgen.structure.templatesystem.RuleProcessor;
-import net.minecraft.world.level.levelgen.structure.templatesystem.StructureProcessorList;
-import net.minecraft.world.level.levelgen.structure.templatesystem.rule.blockentity.AppendLoot;
 import net.minecraft.world.level.material.Fluids;
-import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 
 public class MiscOverworldFeatures {
     public static final ResourceKey<Feature> ICE_SPIKE = FeatureUtils.createKey("ice_spike");
@@ -88,7 +_,7 @@
         context.register(
             ICE_PATCH,
             new DiskFeature(
-                BlockStateProvider.holderOf(Blocks.PACKED_ICE),
+                BlockStateProvider.simple(Blocks.PACKED_ICE),
                 BlockPredicate.matchesBlocks(Blocks.DIRT, Blocks.GRASS_BLOCK, Blocks.PODZOL, Blocks.COARSE_DIRT, Blocks.MYCELIUM, Blocks.SNOW_BLOCK, Blocks.ICE),
                 UniformInt.of(2, 3),
                 1
@@ -103,8 +_,8 @@
         context.register(
             LAKE_LAVA,
             new LakeFeature(
-                BlockStateProvider.holderOf(Blocks.LAVA),
-                BlockStateProvider.holderOf(Blocks.STONE),
+                BlockStateProvider.simple(Blocks.LAVA.defaultBlockState()),
+                BlockStateProvider.simple(Blocks.STONE.defaultBlockState()),
                 BlockPredicate.alwaysTrue(),
                 BlockPredicate.not(BlockPredicate.matchesTag(BlockTags.FEATURES_CANNOT_REPLACE)),
                 BlockPredicate.not(BlockPredicate.matchesTag(BlockTags.LAVA_POOL_STONE_CANNOT_REPLACE))
@@ -116,8 +_,8 @@
                 HolderSet.direct(
                     PlacementUtils.inlinePlaced(
                         new LakeFeature(
-                            BlockStateProvider.holderOf(Blocks.WATER),
-                            BlockStateProvider.holderOf(Blocks.SULFUR),
+                            BlockStateProvider.simple(Blocks.WATER.defaultBlockState()),
+                            BlockStateProvider.simple(Blocks.SULFUR.defaultBlockState()),
                             BlockPredicate.not(BlockPredicate.matchesBlocks(Blocks.SULFUR_SPIKE)),
                             BlockPredicate.not(BlockPredicate.matchesTag(BlockTags.FEATURES_CANNOT_REPLACE)),
                             BlockPredicate.not(BlockPredicate.matchesTag(BlockTags.LAVA_POOL_STONE_CANNOT_REPLACE))
@@ -125,7 +_,7 @@
                     ),
                     PlacementUtils.inlinePlaced(
                         new SimpleBlockFeature(
-                            BlockStateProvider.of(Blocks.POTENT_SULFUR.defaultBlockState().setValue(PotentSulfurBlock.STATE, PotentSulfurState.WET))
+                            BlockStateProvider.simple(Blocks.POTENT_SULFUR.defaultBlockState().setValue(PotentSulfurBlock.STATE, PotentSulfurState.WET))
                         ),
                         EnvironmentScanPlacement.scanningFor(
                             Direction.DOWN, BlockPredicate.allOf(BlockPredicate.solid(), BlockPredicate.matchesFluids(Direction.UP, Fluids.WATER)), 4
@@ -135,7 +_,7 @@
             )
         );
         BiFunction<Integer, Integer, Holder<PlacedFeature>> tuffCover = (count, spread) -> PlacementUtils.inlinePlaced(
-            new SimpleBlockFeature(BlockStateProvider.of(Blocks.TUFF)),
+            new SimpleBlockFeature(BlockStateProvider.simple(Blocks.TUFF)),
             CountPlacement.of(count),
             OffsetPlacement.ofTriangle(spread, 3),
             EnvironmentScanPlacement.scanningFor(Direction.DOWN, BlockPredicate.solid(), 4),
@@ -216,24 +_,19 @@
             )
         );
         context.register(
-            DISK_CLAY,
-            new DiskFeature(BlockStateProvider.holderOf(Blocks.CLAY), BlockPredicate.matchesBlocks(Blocks.DIRT, Blocks.CLAY), UniformInt.of(2, 3), 1)
+            DISK_CLAY, new DiskFeature(BlockStateProvider.simple(Blocks.CLAY), BlockPredicate.matchesBlocks(Blocks.DIRT, Blocks.CLAY), UniformInt.of(2, 3), 1)
         );
         context.register(
             DISK_GRAVEL,
-            new DiskFeature(BlockStateProvider.holderOf(Blocks.GRAVEL), BlockPredicate.matchesBlocks(Blocks.DIRT, Blocks.GRASS_BLOCK), UniformInt.of(2, 5), 2)
+            new DiskFeature(BlockStateProvider.simple(Blocks.GRAVEL), BlockPredicate.matchesBlocks(Blocks.DIRT, Blocks.GRASS_BLOCK), UniformInt.of(2, 5), 2)
         );
         context.register(
             DISK_SAND,
             new DiskFeature(
-                Holder.direct(
-                    new RuleBasedStateProvider(
-                        BlockStateProvider.holderOf(Blocks.SAND),
-                        List.of(
-                            new RuleBasedStateProvider.Rule(
-                                BlockPredicate.matchesBlocks(Direction.DOWN, Blocks.AIR), BlockStateProvider.holderOf(Blocks.SANDSTONE)
-                            )
-                        )
+                new RuleBasedStateProvider(
+                    BlockStateProvider.simple(Blocks.SAND),
+                    List.of(
+                        new RuleBasedStateProvider.Rule(BlockPredicate.matchesBlocks(Direction.DOWN, Blocks.AIR), BlockStateProvider.simple(Blocks.SANDSTONE))
                     )
                 ),
                 BlockPredicate.matchesBlocks(Blocks.DIRT, Blocks.GRASS_BLOCK),
@@ -245,16 +_,14 @@
         context.register(
             DISK_GRASS,
             new DiskFeature(
-                Holder.direct(
-                    new RuleBasedStateProvider(
-                        BlockStateProvider.holderOf(Blocks.DIRT),
-                        List.of(
-                            new RuleBasedStateProvider.Rule(
-                                BlockPredicate.not(
-                                    BlockPredicate.anyOf(BlockPredicate.solid(Direction.UP), BlockPredicate.matchesFluids(Direction.UP, Fluids.WATER))
-                                ),
-                                BlockStateProvider.holderOf(Blocks.GRASS_BLOCK)
-                            )
+                new RuleBasedStateProvider(
+                    BlockStateProvider.simple(Blocks.DIRT),
+                    List.of(
+                        new RuleBasedStateProvider.Rule(
+                            BlockPredicate.not(
+                                BlockPredicate.anyOf(BlockPredicate.solid(Direction.UP), BlockPredicate.matchesFluids(Direction.UP, Fluids.WATER))
+                            ),
+                            BlockStateProvider.simple(Blocks.GRASS_BLOCK)
                         )
                     )
                 ),
@@ -265,49 +_,7 @@
         );
         context.register(BONUS_CHEST, new BonusChestFeature());
         context.register(VOID_START_PLATFORM, new VoidStartPlatformFeature());
-        StructureProcessorList wellSuspiciousSandProcessor = new StructureProcessorList(
-            List.of(
-                new RuleProcessor(
-                    List.of(
-                        new ProcessorRule(
-                            new BlockMatchTest(Blocks.SUSPICIOUS_SAND),
-                            AlwaysTrueTest.INSTANCE,
-                            PosAlwaysTrueTest.INSTANCE,
-                            Blocks.SUSPICIOUS_SAND.defaultBlockState(),
-                            new AppendLoot(BuiltInLootTables.DESERT_WELL_ARCHAEOLOGY)
-                        )
-                    )
-                )
-            )
-        );
-        context.register(
-            DESERT_WELL,
-            new OverlayFeature(
-                HolderSet.direct(
-                    PlacementUtils.inlinePlaced(TemplateFeature.simple(Identifier.withDefaultNamespace("desert_well/well")), OffsetPlacement.of(0, -2, 0)),
-                    PlacementUtils.inlinePlaced(
-                        TemplateFeature.simple(Identifier.withDefaultNamespace("desert_well/suspicious_sand"), Holder.direct(wellSuspiciousSandProcessor)),
-                        new RandomlySelectedPlacement(
-                            OffsetPlacement.of(0, -1, 0),
-                            OffsetPlacement.of(-1, -1, 0),
-                            OffsetPlacement.of(0, -1, 1),
-                            OffsetPlacement.of(1, -1, 0),
-                            OffsetPlacement.of(0, -1, -1)
-                        )
-                    ),
-                    PlacementUtils.inlinePlaced(
-                        TemplateFeature.simple(Identifier.withDefaultNamespace("desert_well/suspicious_sand"), Holder.direct(wellSuspiciousSandProcessor)),
-                        new RandomlySelectedPlacement(
-                            OffsetPlacement.of(0, -2, 0),
-                            OffsetPlacement.of(-1, -2, 0),
-                            OffsetPlacement.of(0, -2, 1),
-                            OffsetPlacement.of(1, -2, 0),
-                            OffsetPlacement.of(0, -2, -1)
-                        )
-                    )
-                )
-            )
-        );
+        context.register(DESERT_WELL, new DesertWellFeature());
         context.register(
             SPRING_LAVA_OVERWORLD,
             new SpringFeature(
