--- a/net/minecraft/data/worldgen/features/NetherFeatures.java
+++ b/net/minecraft/data/worldgen/features/NetherFeatures.java
@@ -88,7 +_,7 @@
                     new Weighted<>(
                         PlacementUtils.inlinePlaced(
                             new SteppedColumnClusterFeature(
-                                BlockStateProvider.holderOf(Blocks.BASALT),
+                                BlockStateProvider.simple(Blocks.BASALT),
                                 BlockPredicate.matchesBlocks(Blocks.BASALT),
                                 replacedByBasaltColumns,
                                 blocks.getOrThrow(BlockTags.CANNOT_PLACE_BASALT_PILLAR_ON),
@@ -103,7 +_,7 @@
                     new Weighted<>(
                         PlacementUtils.inlinePlaced(
                             new SteppedColumnClusterFeature(
-                                BlockStateProvider.holderOf(Blocks.BASALT),
+                                BlockStateProvider.simple(Blocks.BASALT),
                                 BlockPredicate.matchesBlocks(Blocks.BASALT),
                                 replacedByBasaltColumns,
                                 blocks.getOrThrow(BlockTags.CANNOT_PLACE_BASALT_PILLAR_ON),
@@ -125,7 +_,7 @@
                     new Weighted<>(
                         PlacementUtils.inlinePlaced(
                             new SteppedColumnClusterFeature(
-                                BlockStateProvider.holderOf(Blocks.BASALT),
+                                BlockStateProvider.simple(Blocks.BASALT),
                                 BlockPredicate.matchesBlocks(Blocks.BASALT),
                                 replacedByBasaltColumns,
                                 blocks.getOrThrow(BlockTags.CANNOT_PLACE_BASALT_PILLAR_ON),
@@ -140,7 +_,7 @@
                     new Weighted<>(
                         PlacementUtils.inlinePlaced(
                             new SteppedColumnClusterFeature(
-                                BlockStateProvider.holderOf(Blocks.BASALT),
+                                BlockStateProvider.simple(Blocks.BASALT),
                                 BlockPredicate.matchesBlocks(Blocks.BASALT),
                                 replacedByBasaltColumns,
                                 blocks.getOrThrow(BlockTags.CANNOT_PLACE_BASALT_PILLAR_ON),
@@ -162,7 +_,7 @@
         context.register(
             GLOWSTONE_EXTRA,
             new RandomNeighborSpreadFeature(
-                BlockStateProvider.holderOf(Blocks.GLOWSTONE),
+                BlockStateProvider.simple(Blocks.GLOWSTONE),
                 HolderSet.direct(blocks.getOrThrow(BlockItemIds.GLOWSTONE.block())),
                 BlockPredicate.ONLY_IN_AIR_PREDICATE,
                 ConstantInt.of(1500),
@@ -192,7 +_,7 @@
             )
         );
         context.register(WARPED_FOREST_VEGETION, warpedVegetation);
-        SimpleBlockFeature netherSprouts = new SimpleBlockFeature(BlockStateProvider.of(Blocks.NETHER_SPROUTS));
+        SimpleBlockFeature netherSprouts = new SimpleBlockFeature(BlockStateProvider.simple(Blocks.NETHER_SPROUTS));
         context.register(NETHER_SPROUTS, netherSprouts);
         Feature crimsonNyliumBonemeal = new OverlayFeature(HolderSet.direct(PlacementUtils.inlinePlaced(crimsonVegetation, netherForestBonemealSpread)));
         Feature warpedNyliumBonemeal = new OverlayFeature(
@@ -235,7 +_,7 @@
                 HolderSet.direct(
                     PlacementUtils.inlinePlaced(
                         new RandomNeighborSpreadFeature(
-                            BlockStateProvider.holderOf(Blocks.NETHER_WART_BLOCK),
+                            BlockStateProvider.simple(Blocks.NETHER_WART_BLOCK.defaultBlockState()),
                             HolderSet.direct(blocks.getOrThrow(BlockItemIds.NETHERRACK.block()), blocks.getOrThrow(BlockItemIds.NETHER_WART_BLOCK.block())),
                             BlockPredicate.ONLY_IN_AIR_PREDICATE,
                             ConstantInt.of(200),
@@ -256,14 +_,14 @@
                 )
             )
         );
-        context.register(CRIMSON_ROOTS, new SimpleBlockFeature(BlockStateProvider.of(Blocks.CRIMSON_ROOTS)));
+        context.register(CRIMSON_ROOTS, new SimpleBlockFeature(BlockStateProvider.simple(Blocks.CRIMSON_ROOTS)));
         context.register(
             BASALT_PILLAR,
             new OverlayFeature(
                 HolderSet.direct(
                     PlacementUtils.inlinePlaced(
                         new SingleBlockPillarFeature(
-                            BlockStateProvider.holderOf(Blocks.BASALT),
+                            BlockStateProvider.simple(Blocks.BASALT),
                             BlockPredicate.ONLY_IN_AIR_PREDICATE,
                             Direction.DOWN,
                             1.0F,
@@ -273,10 +_,8 @@
                                         HolderSet.direct(
                                             PlacementUtils.inlinePlaced(
                                                 new ProjectedRandomPatchySquare(
-                                                    Holder.direct(
-                                                        RuleBasedStateProvider.ifTrueThenProvide(
-                                                            BlockPredicate.not(BlockPredicate.matchesTag(Direction.DOWN, BlockTags.AIR)), Blocks.BASALT
-                                                        )
+                                                    RuleBasedStateProvider.ifTrueThenProvide(
+                                                        BlockPredicate.not(BlockPredicate.matchesTag(Direction.DOWN, BlockTags.AIR)), Blocks.BASALT
                                                     ),
                                                     BlockPredicate.ONLY_IN_AIR_PREDICATE,
                                                     ConstantInt.of(3),
@@ -285,22 +_,22 @@
                                                 OffsetPlacement.of(0, -1, 0)
                                             ),
                                             PlacementUtils.inlinePlaced(
-                                                new SimpleBlockFeature(BlockStateProvider.of(Blocks.BASALT)),
+                                                new SimpleBlockFeature(BlockStateProvider.simple(Blocks.BASALT)),
                                                 RarityFilter.onAverageOnceEvery(2),
                                                 OffsetPlacement.of(1, 0, 0)
                                             ),
                                             PlacementUtils.inlinePlaced(
-                                                new SimpleBlockFeature(BlockStateProvider.of(Blocks.BASALT)),
+                                                new SimpleBlockFeature(BlockStateProvider.simple(Blocks.BASALT)),
                                                 RarityFilter.onAverageOnceEvery(2),
                                                 OffsetPlacement.of(-1, 0, 0)
                                             ),
                                             PlacementUtils.inlinePlaced(
-                                                new SimpleBlockFeature(BlockStateProvider.of(Blocks.BASALT)),
+                                                new SimpleBlockFeature(BlockStateProvider.simple(Blocks.BASALT)),
                                                 RarityFilter.onAverageOnceEvery(2),
                                                 OffsetPlacement.of(0, 0, 1)
                                             ),
                                             PlacementUtils.inlinePlaced(
-                                                new SimpleBlockFeature(BlockStateProvider.of(Blocks.BASALT)),
+                                                new SimpleBlockFeature(BlockStateProvider.simple(Blocks.BASALT)),
                                                 RarityFilter.onAverageOnceEvery(2),
                                                 OffsetPlacement.of(0, 0, -1)
                                             )
@@ -311,19 +_,19 @@
                         )
                     ),
                     PlacementUtils.inlinePlaced(
-                        new SingleBlockPillarFeature(BlockStateProvider.holderOf(Blocks.BASALT), BlockPredicate.ONLY_IN_AIR_PREDICATE, Direction.DOWN, 0.9F),
+                        new SingleBlockPillarFeature(BlockStateProvider.simple(Blocks.BASALT), BlockPredicate.ONLY_IN_AIR_PREDICATE, Direction.DOWN, 0.9F),
                         OffsetPlacement.of(1, 0, 0)
                     ),
                     PlacementUtils.inlinePlaced(
-                        new SingleBlockPillarFeature(BlockStateProvider.holderOf(Blocks.BASALT), BlockPredicate.ONLY_IN_AIR_PREDICATE, Direction.DOWN, 0.9F),
+                        new SingleBlockPillarFeature(BlockStateProvider.simple(Blocks.BASALT), BlockPredicate.ONLY_IN_AIR_PREDICATE, Direction.DOWN, 0.9F),
                         OffsetPlacement.of(-1, 0, 0)
                     ),
                     PlacementUtils.inlinePlaced(
-                        new SingleBlockPillarFeature(BlockStateProvider.holderOf(Blocks.BASALT), BlockPredicate.ONLY_IN_AIR_PREDICATE, Direction.DOWN, 0.9F),
+                        new SingleBlockPillarFeature(BlockStateProvider.simple(Blocks.BASALT), BlockPredicate.ONLY_IN_AIR_PREDICATE, Direction.DOWN, 0.9F),
                         OffsetPlacement.of(0, 0, 1)
                     ),
                     PlacementUtils.inlinePlaced(
-                        new SingleBlockPillarFeature(BlockStateProvider.holderOf(Blocks.BASALT), BlockPredicate.ONLY_IN_AIR_PREDICATE, Direction.DOWN, 0.9F),
+                        new SingleBlockPillarFeature(BlockStateProvider.simple(Blocks.BASALT), BlockPredicate.ONLY_IN_AIR_PREDICATE, Direction.DOWN, 0.9F),
                         OffsetPlacement.of(0, 0, -1)
                     )
                 )
@@ -347,8 +_,8 @@
             SPRING_NETHER_OPEN,
             new SpringFeature(Fluids.LAVA.defaultFluidState(), false, 4, 1, HolderSet.direct(Block::builtInRegistryHolder, Blocks.NETHERRACK))
         );
-        context.register(FIRE, new SimpleBlockFeature(BlockStateProvider.of(Blocks.FIRE)));
-        context.register(SOUL_FIRE, new SimpleBlockFeature(BlockStateProvider.of(Blocks.SOUL_FIRE)));
+        context.register(FIRE, new SimpleBlockFeature(BlockStateProvider.simple(Blocks.FIRE)));
+        context.register(SOUL_FIRE, new SimpleBlockFeature(BlockStateProvider.simple(Blocks.SOUL_FIRE)));
     }
 
     private static Feature createVines(int minHeight, int maxHeight, Direction direction, Block mainBlock, Block tipBlock) {
@@ -361,9 +_,9 @@
         );
         return new BlockColumnFeature(
             List.of(
-                BlockColumnFeature.layer(mainHeight, BlockStateProvider.of(mainBlock)),
+                BlockColumnFeature.layer(mainHeight, BlockStateProvider.simple(mainBlock)),
                 BlockColumnFeature.layer(
-                    ConstantInt.of(1), new RandomizedIntStateProvider(BlockStateProvider.of(tipBlock), GrowingPlantHeadBlock.AGE, UniformInt.of(17, 25))
+                    ConstantInt.of(1), new RandomizedIntStateProvider(BlockStateProvider.simple(tipBlock), GrowingPlantHeadBlock.AGE, UniformInt.of(17, 25))
                 )
             ),
             direction,
