--- a/net/minecraft/data/worldgen/features/PileFeatures.java
+++ b/net/minecraft/data/worldgen/features/PileFeatures.java
@@ -1,6 +_,5 @@
 package net.minecraft.data.worldgen.features;
 
-import net.minecraft.core.Holder;
 import net.minecraft.data.worldgen.BootstrapContext;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.util.random.WeightedList;
@@ -20,26 +_,22 @@
     public static final ResourceKey<Feature> PILE_PUMPKIN = FeatureUtils.createKey("pile_pumpkin");
 
     public static void bootstrap(BootstrapContext<Feature> context) {
-        context.register(PILE_HAY, new BlockPileFeature(Holder.direct(new RotatedBlockProvider(BlockStateProvider.of(Blocks.HAY_BLOCK)))));
-        context.register(PILE_MELON, new BlockPileFeature(BlockStateProvider.holderOf(Blocks.MELON)));
-        context.register(PILE_SNOW, new BlockPileFeature(BlockStateProvider.holderOf(Blocks.SNOW)));
+        context.register(PILE_HAY, new BlockPileFeature(new RotatedBlockProvider(BlockStateProvider.simple(Blocks.HAY_BLOCK))));
+        context.register(PILE_MELON, new BlockPileFeature(BlockStateProvider.simple(Blocks.MELON)));
+        context.register(PILE_SNOW, new BlockPileFeature(BlockStateProvider.simple(Blocks.SNOW)));
         context.register(
             PILE_ICE,
             new BlockPileFeature(
-                Holder.direct(
-                    new WeightedStateProvider(
-                        WeightedList.<BlockState>builder().add(Blocks.BLUE_ICE.defaultBlockState(), 1).add(Blocks.PACKED_ICE.defaultBlockState(), 5)
-                    )
+                new WeightedStateProvider(
+                    WeightedList.<BlockState>builder().add(Blocks.BLUE_ICE.defaultBlockState(), 1).add(Blocks.PACKED_ICE.defaultBlockState(), 5)
                 )
             )
         );
         context.register(
             PILE_PUMPKIN,
             new BlockPileFeature(
-                Holder.direct(
-                    new WeightedStateProvider(
-                        WeightedList.<BlockState>builder().add(Blocks.PUMPKIN.defaultBlockState(), 19).add(Blocks.JACK_O_LANTERN.defaultBlockState(), 1)
-                    )
+                new WeightedStateProvider(
+                    WeightedList.<BlockState>builder().add(Blocks.PUMPKIN.defaultBlockState(), 19).add(Blocks.JACK_O_LANTERN.defaultBlockState(), 1)
                 )
             )
         );
