--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -14,6 +14,10 @@
 import net.minecraft.world.phys.EntityHitResult;
 import net.minecraft.world.phys.HitResult;
 
+// CraftBukkit start
+import org.bukkit.event.entity.EntityRemoveEvent;
+// CraftBukkit end
+
 public class Snowball extends ThrowableItemProjectile {
 
     public Snowball(EntityType<? extends Snowball> entitytype, Level level) {
@@ -65,7 +69,7 @@
         super.onHit(hitresult);
         if (!this.level().isClientSide()) {
             this.level().broadcastEntityEvent(this, (byte) 3);
-            this.discard();
+            this.discard(EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
 
     }
