--- a/net/minecraft/client/gui/Font.java
+++ b/net/minecraft/client/gui/Font.java
@@ -82,10 +_,6 @@
         return outlineOutput;
     }
 
-    public TextRenderable prepareBackground(float x0, float y0, float x1, float y1, int color) {
-        return this.provider.effect().createEffect(x0, y0, x1, y1, -0.01F, color, 0, 0.0F);
-    }
-
     private BakedGlyph getGlyph(int codepoint, Style style) {
         GlyphSource glyphSource = this.getGlyphSource(style.getFont());
         BakedGlyph glyph = glyphSource.getGlyph(codepoint);
