--- a/net/minecraft/client/gui/Font.java
+++ b/net/minecraft/client/gui/Font.java
@@ -27,10 +_,8 @@
 import net.neoforged.api.distmarker.OnlyIn;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.extensions.IFontExtension;
-
 @OnlyIn(Dist.CLIENT)
-public class Font implements IFontExtension {
+public class Font {
     private static final float EFFECT_DEPTH = 0.01F;
     private static final float OVER_EFFECT_DEPTH = 0.01F;
     private static final float UNDER_EFFECT_DEPTH = -0.01F;
