--- a/net/minecraft/client/gui/GuiGraphicsExtractor.java
+++ b/net/minecraft/client/gui/GuiGraphicsExtractor.java
@@ -84,10 +_,8 @@
 import org.joml.Vector3fc;
 import org.jspecify.annotations.Nullable;
 
-import net.neoforged.neoforge.client.extensions.GuiGraphicsExtractorExtension;
-
 @OnlyIn(Dist.CLIENT)
-public class GuiGraphicsExtractor implements GuiGraphicsExtractorExtension {
+public class GuiGraphicsExtractor {
     private static final int EXTRA_SPACE_AFTER_FIRST_TOOLTIP_LINE = 2;
     private final Minecraft minecraft;
     private final Matrix3x2fStack pose;
@@ -1166,7 +_,7 @@
         }
     }
 
-    public void componentHoverEffect(Font font, Style hoveredStyle, int xMouse, int yMouse) {
+    private void componentHoverEffect(Font font, Style hoveredStyle, int xMouse, int yMouse) {
         if (hoveredStyle.getHoverEvent() != null) {
             switch (hoveredStyle.getHoverEvent()) {
                 case HoverEvent.ShowItem(ItemStackTemplate item):
