--- a/net/minecraft/client/gui/GuiGraphicsExtractor.java
+++ b/net/minecraft/client/gui/GuiGraphicsExtractor.java
@@ -272,17 +_,15 @@
         this.text(font, text, x - font.width(text) / 2, y, color);
     }
 
-    public int textWithWordWrap(Font font, FormattedText string, int x, int y, int width, int col) {
-        return this.textWithWordWrap(font, string, x, y, width, col, true);
+    public void textWithWordWrap(Font font, FormattedText string, int x, int y, int width, int col) {
+        this.textWithWordWrap(font, string, x, y, width, col, true);
     }
 
-    public int textWithWordWrap(Font font, FormattedText string, int x, int y, int width, int col, boolean dropShadow) {
+    public void textWithWordWrap(Font font, FormattedText string, int x, int y, int width, int col, boolean dropShadow) {
         for (FormattedCharSequence line : font.split(string, width)) {
             this.text(font, line, x, y, col, dropShadow);
             y += 9;
         }
-
-        return y;
     }
 
     public void textWithBackdrop(Font font, Component str, int textX, int textY, int textWidth, int textColor) {
