--- a/net/minecraft/client/gui/Hud.java
+++ b/net/minecraft/client/gui/Hud.java
@@ -774,10 +_,10 @@
             int currentHealth = Mth.ceil(player.getHealth());
             boolean blink = this.healthBlinkTime > this.tickCount && (this.healthBlinkTime - this.tickCount) / 3L % 2L == 1L;
             long timeMillis = Util.getMillis();
-            if (currentHealth < this.lastHealth && player.damageCooldownTime > 0) {
+            if (currentHealth < this.lastHealth && player.invulnerableTime > 0) {
                 this.lastHealthTime = timeMillis;
                 this.healthBlinkTime = this.tickCount + 20;
-            } else if (currentHealth > this.lastHealth && player.damageCooldownTime > 0) {
+            } else if (currentHealth > this.lastHealth && player.invulnerableTime > 0) {
                 this.lastHealthTime = timeMillis;
                 this.healthBlinkTime = this.tickCount + 10;
             }
