--- a/net/minecraft/world/phys/shapes/CubeVoxelShape.java
+++ b/net/minecraft/world/phys/shapes/CubeVoxelShape.java
@@ -5,8 +_,9 @@
 import net.minecraft.util.Mth;
 
 public final class CubeVoxelShape extends VoxelShape {
-    protected CubeVoxelShape(DiscreteVoxelShape p_82765_) {
+    public CubeVoxelShape(DiscreteVoxelShape p_82765_) {
         super(p_82765_);
+        this.moonrise$initCache(); // Paper - optimise collisions
     }
 
     @Override
