--- a/net/minecraft/world/phys/shapes/DiscreteCubeMerger.java
+++ b/net/minecraft/world/phys/shapes/DiscreteCubeMerger.java
@@ -8,7 +_,7 @@
     private final int firstDiv;
     private final int secondDiv;
 
-    public DiscreteCubeMerger(int p_82776_, int p_82777_) {
+    DiscreteCubeMerger(int p_82776_, int p_82777_) {
         this.result = new CubePointRange((int)Shapes.lcm(p_82776_, p_82777_));
         int i = IntMath.gcd(p_82776_, p_82777_);
         this.firstDiv = p_82776_ / i;
