--- a/net/minecraft/world/phys/shapes/SliceShape.java
+++ b/net/minecraft/world/phys/shapes/SliceShape.java
@@ -12,6 +_,7 @@
         super(makeSlice(delegate.shape, axis, point));
         this.delegate = delegate;
         this.axis = axis;
+        ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)this).moonrise$initCache(); // Paper - optimise collisions
     }
 
     private static DiscreteVoxelShape makeSlice(DiscreteVoxelShape delegate, Direction.Axis axis, int point) {
