--- a/net/minecraft/world/phys/shapes/VoxelShape.java
+++ b/net/minecraft/world/phys/shapes/VoxelShape.java
@@ -19,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 public abstract class VoxelShape {
-    protected final DiscreteVoxelShape shape;
+    public final DiscreteVoxelShape shape;
     private @Nullable VoxelShape @Nullable [] faces;
 
     protected VoxelShape(DiscreteVoxelShape shape) {
