--- a/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
+++ b/net/minecraft/world/entity/monster/piglin/AbstractPiglin.java
@@ -103,7 +103,7 @@
     protected void finishConversion(ServerLevel level) {
         this.convertTo(EntityType.ZOMBIFIED_PIGLIN, ConversionParams.single(this, true, true), (zombifiedpiglin) -> {
             zombifiedpiglin.addEffect(new MobEffectInstance(MobEffects.NAUSEA, 200, 0));
-        });
+        }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.PIGLIN_ZOMBIFIED, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PIGLIN_ZOMBIFIED); // CraftBukkit - add spawn and transform reasons
     }
 
     public boolean isAdult() {
