--- a/net/minecraft/client/model/monster/skeleton/SkeletonModel.java
+++ b/net/minecraft/client/model/monster/skeleton/SkeletonModel.java
@@ -102,17 +_,17 @@
     public void setupAnim(S state) {
         super.setupAnim(state);
         if (state.isAggressive && !state.isHoldingBow) {
-            float swingAnimation = state.swingAnimation;
-            float swing2 = Mth.sin(swingAnimation * (float) Math.PI);
-            float swing = Mth.sin((1.0F - (1.0F - swingAnimation) * (1.0F - swingAnimation)) * (float) Math.PI);
+            float attackTime = state.attackTime;
+            float attack2 = Mth.sin(attackTime * (float) Math.PI);
+            float attack = Mth.sin((1.0F - (1.0F - attackTime) * (1.0F - attackTime)) * (float) Math.PI);
             this.rightArm.zRot = 0.0F;
             this.leftArm.zRot = 0.0F;
-            this.rightArm.yRot = -(0.1F - swing2 * 0.6F);
-            this.leftArm.yRot = 0.1F - swing2 * 0.6F;
+            this.rightArm.yRot = -(0.1F - attack2 * 0.6F);
+            this.leftArm.yRot = 0.1F - attack2 * 0.6F;
             this.rightArm.xRot = (float) (-Math.PI / 2);
             this.leftArm.xRot = (float) (-Math.PI / 2);
-            this.rightArm.xRot -= swing2 * 1.2F - swing * 0.4F;
-            this.leftArm.xRot -= swing2 * 1.2F - swing * 0.4F;
+            this.rightArm.xRot -= attack2 * 1.2F - attack * 0.4F;
+            this.leftArm.xRot -= attack2 * 1.2F - attack * 0.4F;
             AnimationUtils.bobArms(this.rightArm, this.leftArm, state.ageInTicks);
         }
     }
