--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -84,6 +84,11 @@
 import net.minecraft.world.phys.shapes.VoxelShape;
 import org.jspecify.annotations.Nullable;
 
+// CraftBukkit start
+import net.minecraft.world.item.context.UseOnContext;
+import net.minecraft.world.level.ServerExplosion;
+// CraftBukkit end
+
 public abstract class BlockBehaviour implements FeatureElement {
 
     protected static final Direction[] UPDATE_SHAPE_ORDER = new Direction[]{Direction.WEST, Direction.EAST, Direction.NORTH, Direction.SOUTH, Direction.DOWN, Direction.UP};
@@ -158,6 +163,12 @@
 
     protected void onPlace(BlockState blockstate, Level level, BlockPos blockpos, BlockState blockstate1, boolean flag) {}
 
+    // CraftBukkit start
+    protected void onPlace(BlockState blockstate, Level level, BlockPos blockpos, BlockState blockstate1, boolean flag, @Nullable UseOnContext context) {
+        this.onPlace(blockstate, level, blockpos, blockstate1, flag);
+    }
+    // CraftBukkit end
+
     protected void affectNeighborsAfterRemoval(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, boolean flag) {}
 
     protected void onExplosionHit(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, Explosion explosion, BiConsumer<ItemStack, BlockPos> biconsumer) {
@@ -169,8 +180,10 @@
                 BlockEntity blockentity = blockstate.hasBlockEntity() ? serverlevel.getBlockEntity(blockpos) : null;
                 LootParams.Builder lootparams_builder = (new LootParams.Builder(serverlevel)).withParameter(LootContextParams.ORIGIN, Vec3.atCenterOf(blockpos)).withParameter(LootContextParams.TOOL, ItemStack.EMPTY).withOptionalParameter(LootContextParams.BLOCK_ENTITY, blockentity).withOptionalParameter(LootContextParams.THIS_ENTITY, explosion.getDirectSourceEntity());
 
-                if (explosion.getBlockInteraction() == Explosion.BlockInteraction.DESTROY_WITH_DECAY) {
-                    lootparams_builder.withParameter(LootContextParams.EXPLOSION_RADIUS, explosion.radius());
+                // CraftBukkit start - add yield
+                if (explosion instanceof ServerExplosion serverExplosion && serverExplosion.yield < 1.0F) {
+                    lootparams_builder.withParameter(LootContextParams.EXPLOSION_RADIUS, 1.0F / serverExplosion.yield);
+                    // CraftBukkit end
                 }
 
                 blockstate.spawnAfterBreak(serverlevel, blockpos, ItemStack.EMPTY, flag);
@@ -1103,7 +1116,13 @@
         }
 
         public void onPlace(Level level, BlockPos blockpos, BlockState blockstate, boolean flag) {
-            this.getBlock().onPlace(this.asState(), level, blockpos, blockstate, flag);
+            // CraftBukkit start
+            this.onPlace(level, blockpos, blockstate, flag, null);
+        }
+
+        public void onPlace(Level level, BlockPos blockpos, BlockState blockstate, boolean flag, @Nullable UseOnContext context) {
+            this.getBlock().onPlace(this.asState(), level, blockpos, blockstate, flag, context);
+            // CraftBukkit end
         }
 
         public void affectNeighborsAfterRemoval(ServerLevel serverlevel, BlockPos blockpos, boolean flag) {
