--- a/net/minecraft/data/worldgen/BootstrapContext.java
+++ b/net/minecraft/data/worldgen/BootstrapContext.java
@@ -1,8 +_,19 @@
 package net.minecraft.data.worldgen;
 
+import com.mojang.serialization.Lifecycle;
 import net.minecraft.core.Holder;
+import net.minecraft.core.HolderGetter;
+import net.minecraft.core.Registry;
 import net.minecraft.resources.ResourceKey;
 
-public interface BootstrapContext<T> extends BootstrapContextAccess {
-    Holder.Reference<T> register(ResourceKey<T> key, T value);
+public interface BootstrapContext<T> {
+    Holder.Reference<T> register(ResourceKey<T> key, T value, Lifecycle lifecycle);
+
+    default Holder.Reference<T> register(ResourceKey<T> key, T value) {
+        return this.register(key, value, Lifecycle.stable());
+    }
+
+    <S> HolderGetter<S> lookup(ResourceKey<? extends Registry<? extends S>> key);
+
+    default <S> java.util.Optional<net.minecraft.core.HolderLookup<S>> holderLookup(ResourceKey<? extends Registry<? extends S>> registry) { return java.util.Optional.empty(); }
 }
