--- a/net/minecraft/data/worldgen/NoiseData.java
+++ b/net/minecraft/data/worldgen/NoiseData.java
@@ -6,9 +_,9 @@
 
 public class NoiseData {
     @Deprecated
-    public static final NormalNoise DEFAULT_SHIFT = NormalNoise.createParity(-3, 1.0, 1.0, 1.0, 0.0);
+    public static final NormalNoise.NoiseParameters DEFAULT_SHIFT = new NormalNoise.NoiseParameters(-3, 1.0, 1.0, 1.0, 0.0);
 
-    public static void bootstrap(BootstrapContext<NormalNoise> context) {
+    public static void bootstrap(BootstrapContext<NormalNoise.NoiseParameters> context) {
         registerBiomeNoises(context, 0, Noises.TEMPERATURE, Noises.VEGETATION, Noises.CONTINENTALNESS, Noises.EROSION);
         registerBiomeNoises(context, -2, Noises.TEMPERATURE_LARGE, Noises.VEGETATION_LARGE, Noises.CONTINENTALNESS_LARGE, Noises.EROSION_LARGE);
         register(context, Noises.TEMPERATURE_NETHER, -7, 1.0, 1.0);
@@ -70,12 +_,12 @@
     }
 
     private static void registerBiomeNoises(
-        BootstrapContext<NormalNoise> context,
+        BootstrapContext<NormalNoise.NoiseParameters> context,
         int octaveOffset,
-        ResourceKey<NormalNoise> temperature,
-        ResourceKey<NormalNoise> vegetation,
-        ResourceKey<NormalNoise> continentalness,
-        ResourceKey<NormalNoise> erosion
+        ResourceKey<NormalNoise.NoiseParameters> temperature,
+        ResourceKey<NormalNoise.NoiseParameters> vegetation,
+        ResourceKey<NormalNoise.NoiseParameters> continentalness,
+        ResourceKey<NormalNoise.NoiseParameters> erosion
     ) {
         register(context, temperature, -10 + octaveOffset, 1.5, 0.0, 1.0, 0.0, 0.0, 0.0);
         register(context, vegetation, -8 + octaveOffset, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0);
@@ -83,7 +_,13 @@
         register(context, erosion, -9 + octaveOffset, 1.0, 1.0, 0.0, 1.0, 1.0);
     }
 
-    private static void register(BootstrapContext<NormalNoise> context, ResourceKey<NormalNoise> key, int firstOctave, double... amplitudes) {
-        context.register(key, NormalNoise.createParity(firstOctave, amplitudes));
+    private static void register(
+        BootstrapContext<NormalNoise.NoiseParameters> context,
+        ResourceKey<NormalNoise.NoiseParameters> key,
+        int firstOctave,
+        double firstAmplitude,
+        double... amplitudes
+    ) {
+        context.register(key, new NormalNoise.NoiseParameters(firstOctave, firstAmplitude, amplitudes));
     }
 }
