--- a/net/minecraft/data/worldgen/StructureSets.java
+++ b/net/minecraft/data/worldgen/StructureSets.java
@@ -220,7 +_,7 @@
                     StructureSet.entry(structures.getOrThrow(BuiltinStructures.ABANDONED_CAMP_WINDSWEPT_FOREST)),
                     StructureSet.entry(structures.getOrThrow(BuiltinStructures.ABANDONED_CAMP_WOODED_BADLANDS))
                 ),
-                new RandomSpreadStructurePlacement(37, 8, RandomSpreadType.LINEAR, 91231127)
+                new RandomSpreadStructurePlacement(34, 8, RandomSpreadType.LINEAR, 91231127)
             )
         );
     }
