--- a/net/minecraft/world/level/lighting/LeveledPriorityQueue.java
+++ b/net/minecraft/world/level/lighting/LeveledPriorityQueue.java
@@ -26,6 +_,9 @@
     }
 
     public long removeFirstLong() {
+        if (this.isEmpty()) {
+            return 0L;
+        }
         LongLinkedOpenHashSet longlinkedopenhashset = this.queues[this.firstQueuedLevel];
         long i = longlinkedopenhashset.removeFirstLong();
         if (longlinkedopenhashset.isEmpty()) {
