--- a/net/minecraft/world/level/lighting/SkyLightSectionStorage.java
+++ b/net/minecraft/world/level/lighting/SkyLightSectionStorage.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.level.lighting;
 
+import com.mohistmc.youer.YouerConfig;
 import it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 import net.minecraft.core.BlockPos;
@@ -24,6 +_,9 @@
     }
 
     protected int getLightValue(long p_164458_, boolean p_164459_) {
+        if (YouerConfig.SkyLight15) {
+            return 15;
+        }
         long i = SectionPos.blockToSection(p_164458_);
         int j = SectionPos.y(i);
         SkyLightSectionStorage.SkyDataLayerStorageMap skylightsectionstorage$skydatalayerstoragemap = p_164459_
