--- a/net/minecraft/world/effect/HealOrHarmMobEffect.java
+++ b/net/minecraft/world/effect/HealOrHarmMobEffect.java
@@ -17,7 +17,7 @@
     @Override
     public boolean applyEffectTick(ServerLevel serverlevel, LivingEntity livingentity, int i) {
         if (this.isHarm == livingentity.isInvertedHealAndHarm()) {
-            livingentity.heal((float) Math.max(4 << i, 0));
+            livingentity.heal((float) Math.max(4 << i, 0), org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
         } else {
             livingentity.hurtServer(serverlevel, livingentity.damageSources().magic(), (float) (6 << i));
         }
@@ -30,7 +30,7 @@
         if (this.isHarm == livingentity.isInvertedHealAndHarm()) {
             int j = (int) (d0 * (double) (4 << i) + 0.5D);
 
-            livingentity.heal((float) j);
+            livingentity.heal((float) j, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
         } else {
             int k = (int) (d0 * (double) (6 << i) + 0.5D);
 
