--- a/net/minecraft/world/effect/HungerMobEffect.java
+++ b/net/minecraft/world/effect/HungerMobEffect.java
@@ -13,7 +13,7 @@
     @Override
     public boolean applyEffectTick(ServerLevel serverlevel, LivingEntity livingentity, int i) {
         if (livingentity instanceof Player player) {
-            player.causeFoodExhaustion(0.005F * (float) (i + 1));
+            player.causeFoodExhaustion(0.005F * (float) (i + 1), org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.HUNGER_EFFECT); // CraftBukkit - EntityExhaustionEvent
         }
 
         return true;
