--- a/net/minecraft/world/effect/MobEffectUtil.java
+++ b/net/minecraft/world/effect/MobEffectUtil.java
@@ -54,13 +54,19 @@
     }
 
     public static List<ServerPlayer> addEffectToPlayersAround(ServerLevel serverlevel, @Nullable Entity entity, Vec3 vec3, double d0, MobEffectInstance mobeffectinstance, int i) {
+        // CraftBukkit start
+        return addEffectToPlayersAround(serverlevel, entity, vec3, d0, mobeffectinstance, i, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.UNKNOWN);
+    }
+
+    public static List<ServerPlayer> addEffectToPlayersAround(ServerLevel serverlevel, @Nullable Entity entity, Vec3 vec3, double d0, MobEffectInstance mobeffectinstance, int i, org.bukkit.event.entity.EntityPotionEffectEvent.Cause cause) {
+        // CraftBukkit end
         Holder<MobEffect> holder = mobeffectinstance.getEffect();
         List<ServerPlayer> list = serverlevel.getPlayers((serverplayer) -> {
             return serverplayer.gameMode.isSurvival() && (entity == null || !entity.isAlliedTo((Entity) serverplayer)) && vec3.closerThan(serverplayer.position(), d0) && (!serverplayer.hasEffect(holder) || serverplayer.getEffect(holder).getAmplifier() < mobeffectinstance.getAmplifier() || serverplayer.getEffect(holder).endsWithin(i - 1));
         });
 
         list.forEach((serverplayer) -> {
-            serverplayer.addEffect(new MobEffectInstance(mobeffectinstance), entity);
+            serverplayer.addEffect(new MobEffectInstance(mobeffectinstance), entity, cause); // CraftBukkit
         });
         return list;
     }
