--- a/net/minecraft/world/effect/PoisonMobEffect.java
+++ b/net/minecraft/world/effect/PoisonMobEffect.java
@@ -14,7 +14,7 @@
     @Override
     public boolean applyEffectTick(ServerLevel serverlevel, LivingEntity livingentity, int i) {
         if (livingentity.getHealth() > 1.0F) {
-            livingentity.hurtServer(serverlevel, livingentity.damageSources().magic(), 1.0F);
+            livingentity.hurtServer(serverlevel, livingentity.damageSources().poison(), 1.0F);  // CraftBukkit - DamageSource.MAGIC -> CraftEventFactory.POISON
         }
 
         return true;
