--- a/net/minecraft/util/datafix/fixes/BlockStateData.java
+++ b/net/minecraft/util/datafix/fixes/BlockStateData.java
@@ -679,7 +_,7 @@
         return new Dynamic<>(JavaOps.INSTANCE, Map.of("Name", name)).convert(NbtOps.INSTANCE);
     }
 
-    private static Dynamic<?> create(String name, Map<String, String> properties) {
+    public static Dynamic<?> create(String name, Map<String, String> properties) {
         return new Dynamic<>(JavaOps.INSTANCE, Map.of("Name", name, "Properties", properties)).convert(NbtOps.INSTANCE);
     }
 
