--- a/net/minecraft/util/datafix/fixes/ItemSpawnEggFix.java
+++ b/net/minecraft/util/datafix/fixes/ItemSpawnEggFix.java
@@ -17,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 
 public class ItemSpawnEggFix extends DataFix {
-    public static final @Nullable String[] ID_TO_ENTITY = DataFixUtils.make(new String[256], map -> {
+    private static final @Nullable String[] ID_TO_ENTITY = DataFixUtils.make(new String[256], map -> {
         map[1] = "Item";
         map[2] = "XPOrb";
         map[7] = "ThrownEgg";
