--- a/net/minecraft/util/datafix/fixes/ItemSpawnEggFix.java
+++ b/net/minecraft/util/datafix/fixes/ItemSpawnEggFix.java
@@ -16,7 +_,7 @@
 import net.minecraft.util.datafix.schemas.NamespacedSchema;
 
 public class ItemSpawnEggFix extends DataFix {
-    public static final String[] ID_TO_ENTITY = DataFixUtils.make(new String[256], p_341874_ -> {
+    private static final String[] ID_TO_ENTITY = DataFixUtils.make(new String[256], p_341874_ -> {
         p_341874_[1] = "Item";
         p_341874_[2] = "XPOrb";
         p_341874_[7] = "ThrownEgg";
