--- a/net/minecraft/util/datafix/fixes/ItemIdFix.java
+++ b/net/minecraft/util/datafix/fixes/ItemIdFix.java
@@ -14,7 +_,7 @@
 import net.minecraft.util.datafix.schemas.NamespacedSchema;
 
 public class ItemIdFix extends DataFix {
-    public static final Int2ObjectMap<String> ITEM_NAMES = DataFixUtils.make(new Int2ObjectOpenHashMap<>(), map -> {
+    private static final Int2ObjectMap<String> ITEM_NAMES = DataFixUtils.make(new Int2ObjectOpenHashMap<>(), map -> {
         map.put(1, "minecraft:stone");
         map.put(2, "minecraft:grass");
         map.put(3, "minecraft:dirt");
