--- a/net/minecraft/util/datafix/fixes/MapColorComponentRemovalFix.java
+++ /dev/null
@@ -1,16 +_,0 @@
-package net.minecraft.util.datafix.fixes;
-
-import com.mojang.datafixers.schemas.Schema;
-import com.mojang.serialization.Dynamic;
-import org.jspecify.annotations.Nullable;
-
-public class MapColorComponentRemovalFix extends DataComponentRemainderFix {
-    public MapColorComponentRemovalFix(Schema outputSchema) {
-        super(outputSchema, "MapColorComponentRemovalFix", "minecraft:map_color");
-    }
-
-    @Override
-    protected <T> @Nullable Dynamic<T> fixComponent(Dynamic<T> input) {
-        return null;
-    }
-}
