--- a/net/minecraft/util/datafix/fixes/PotDecorationsBlockEntityUnflatteningFix.java
+++ b/net/minecraft/util/datafix/fixes/PotDecorationsBlockEntityUnflatteningFix.java
@@ -3,7 +_,6 @@
 import com.mojang.datafixers.Typed;
 import com.mojang.datafixers.schemas.Schema;
 import com.mojang.datafixers.types.Type;
-import com.mojang.serialization.Dynamic;
 import net.minecraft.util.Util;
 
 public class PotDecorationsBlockEntityUnflatteningFix extends NamedEntityFix {
@@ -14,9 +_,6 @@
     @Override
     protected Typed<?> fix(Typed<?> entity) {
         Type<?> newType = this.getOutputSchema().findChoiceType(References.BLOCK_ENTITY).types().get("minecraft:decorated_pot");
-        return Util.writeAndReadTypedOrThrow(entity, newType, contents -> {
-            Dynamic<?> original = contents.get("sherds").orElseEmptyList();
-            return contents.set("sherds", PotDecorationsComponentUnflatteningFix.unpackList(original));
-        });
+        return Util.writeAndReadTypedOrThrow(entity, newType, contents -> contents.update("sherds", PotDecorationsComponentUnflatteningFix::unpackList));
     }
 }
