--- a/net/minecraft/util/datafix/fixes/PotDecorationsComponentUnflatteningFix.java
+++ b/net/minecraft/util/datafix/fixes/PotDecorationsComponentUnflatteningFix.java
@@ -34,14 +_,8 @@
         List<Optional<String>> decorationIdList = decorationIds.get().map(s -> s.asString().result()).toList();
         Map<Dynamic<T>, Dynamic<T>> result = new HashMap<>(4);
 
-        for (int i = 0; i < 4; i++) {
-            String decorationId;
-            if (i < decorationIdList.size()) {
-                decorationId = decorationIdList.get(i).orElse("minecraft:brick");
-            } else {
-                decorationId = "minecraft:brick";
-            }
-
+        for (int i = 0; i < decorationIdList.size(); i++) {
+            Optional<String> decorationId = decorationIdList.get(i);
             if (decorationId.isEmpty()) {
                 return original;
             }
@@ -50,10 +_,12 @@
                 case 1 -> "left";
                 case 2 -> "right";
                 case 3 -> "front";
-                default -> throw new IndexOutOfBoundsException();
+                default -> null;
             };
-            Map<Dynamic<T>, Dynamic<T>> newStack = Map.of(original.createString("id"), original.createString(decorationId));
-            result.put(original.createString(sideName), original.createMap(newStack));
+            if (sideName != null) {
+                Map<Dynamic<T>, Dynamic<T>> newStack = Map.of(original.createString("id"), original.createString(decorationId.get()));
+                result.put(original.createString(sideName), original.createMap(newStack));
+            }
         }
 
         return original.createMap(result);
