--- a/net/minecraft/server/commands/BossBarCommands.java
+++ b/net/minecraft/server/commands/BossBarCommands.java
@@ -8,7 +_,6 @@
 import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import com.mojang.brigadier.suggestion.SuggestionProvider;
-import com.mojang.brigadier.suggestion.SuggestionsBuilder;
 import java.util.Collection;
 import java.util.Collections;
 import net.minecraft.commands.CommandBuildContext;
