--- a/net/minecraft/server/commands/ChaseCommand.java
+++ b/net/minecraft/server/commands/ChaseCommand.java
@@ -5,7 +_,6 @@
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.arguments.IntegerArgumentType;
 import com.mojang.brigadier.arguments.StringArgumentType;
-import com.mojang.brigadier.context.CommandContext;
 import com.mojang.logging.LogUtils;
 import java.io.IOException;
 import javax.annotation.Nullable;
