--- a/net/minecraft/server/commands/ClearInventoryCommands.java
+++ b/net/minecraft/server/commands/ClearInventoryCommands.java
@@ -72,14 +_,11 @@
 
     private static int clearInventory(CommandSourceStack source, Collection<ServerPlayer> players, Predicate<ItemStack> predicate, int maxCount) throws CommandSyntaxException {
         CommandResponseTracker<ServerPlayer> tracker = CommandResponseTracker.create();
-        boolean countingOnly = maxCount == 0;
 
         for (ServerPlayer player : players) {
-            tracker.track(player, player.getInventory().clearOrCountMatchingItems(predicate, countingOnly, maxCount, player.inventoryMenu.getCraftSlots()));
-            if (!countingOnly) {
-                player.containerMenu.broadcastChanges();
-                player.inventoryMenu.slotsChanged(player.getInventory());
-            }
+            tracker.track(player, player.getInventory().clearOrCountMatchingItems(predicate, maxCount, player.inventoryMenu.getCraftSlots()));
+            player.containerMenu.broadcastChanges();
+            player.inventoryMenu.slotsChanged(player.getInventory());
         }
 
         if (tracker.totalValue() == 0) {
