--- a/net/minecraft/server/commands/DebugConfigCommand.java
+++ b/net/minecraft/server/commands/DebugConfigCommand.java
@@ -2,8 +_,6 @@
 
 import com.mojang.authlib.GameProfile;
 import com.mojang.brigadier.CommandDispatcher;
-import com.mojang.brigadier.context.CommandContext;
-import com.mojang.brigadier.suggestion.SuggestionsBuilder;
 import java.util.HashSet;
 import java.util.Set;
 import java.util.UUID;
