--- a/net/minecraft/server/commands/DebugCommand.java
+++ b/net/minecraft/server/commands/DebugCommand.java
@@ -37,6 +_,7 @@
 import net.minecraft.util.TimeUtil;
 import net.minecraft.util.profiling.ProfileResults;
 import org.apache.commons.io.IOUtils;
+import org.bukkit.command.CommandSender;
 import org.slf4j.Logger;
 
 public class DebugCommand {
@@ -258,6 +_,11 @@
         @Override
         public boolean alwaysAccepts() {
             return true;
+        }
+
+        @Override
+        public CommandSender getBukkitSender(CommandSourceStack wrapper) {
+            return null;
         }
 
         @Override
