--- a/net/minecraft/server/commands/DeOpCommands.java
+++ b/net/minecraft/server/commands/DeOpCommands.java
@@ -41,7 +_,7 @@
             if (playerlist.isOp(gameprofile)) {
                 playerlist.deop(gameprofile);
                 i++;
-                p_136898_.sendSuccess(() -> Component.translatable("commands.deop.success", p_136899_.iterator().next().getName()), true);
+                p_136898_.sendSuccess(() -> Component.translatable("commands.deop.success", gameprofile.getName()), true); // Paper - fixes MC-253721
             }
         }
 
