--- a/net/minecraft/server/commands/EffectCommands.java
+++ b/net/minecraft/server/commands/EffectCommands.java
@@ -29,25 +_,6 @@
     private static final SimpleCommandExceptionType ERROR_CLEAR_SPECIFIC_FAILED = new SimpleCommandExceptionType(
         Component.translatable("commands.effect.clear.specific.failed")
     );
-    private static final CommandResponseTracker.MessagesWithArgs<LivingEntity, MobEffect, Integer> RESPONSE_GIVE = CommandResponseTracker.messages(
-        ERROR_GIVE_FAILED,
-        (entity, var1, effect, duration) -> Component.translatable(
-            "commands.effect.give.success.single", effect.getDisplayName(), entity.getDisplayName(), duration / 20
-        ),
-        (entityCount, var1, effect, duration) -> Component.translatable(
-            "commands.effect.give.success.multiple", effect.getDisplayName(), entityCount, duration / 20
-        )
-    );
-    private static final CommandResponseTracker.Messages<Entity> RESPONSE_CLEAR_ALL = CommandResponseTracker.messages(
-        ERROR_CLEAR_EVERYTHING_FAILED,
-        (entity, var1) -> Component.translatable("commands.effect.clear.everything.success.single", entity.getDisplayName()),
-        (entityCount, var1) -> Component.translatable("commands.effect.clear.everything.success.multiple", entityCount)
-    );
-    private static final CommandResponseTracker.MessagesWithArg<Entity, MobEffect> RESPONSE_CLEAR_SINGLE = CommandResponseTracker.messages(
-        ERROR_CLEAR_SPECIFIC_FAILED,
-        (entity, var1, effect) -> Component.translatable("commands.effect.clear.specific.success.single", effect.getDisplayName(), entity.getDisplayName()),
-        (entityCount, var1, effect) -> Component.translatable("commands.effect.clear.specific.success.multiple", effect.getDisplayName(), entityCount)
-    );
 
     public static void register(CommandDispatcher<CommandSourceStack> dispatcher, CommandBuildContext context) {
         dispatcher.register(
@@ -178,53 +_,100 @@
         boolean particles
     ) throws CommandSyntaxException {
         MobEffect effect = effectHolder.value();
-        int duration = computeDurationInTicks(seconds, effect);
-        CommandResponseTracker<LivingEntity> tracker = CommandResponseTracker.create();
-
-        for (Entity entity : entities) {
-            if (entity instanceof LivingEntity livingEntity) {
-                MobEffectInstance instance = new MobEffectInstance(effectHolder, duration, amplifier, false, particles);
-                tracker.track(livingEntity, livingEntity.addEffect(instance, source.getEntity()));
-            }
-        }
-
-        return tracker.sendFeedback(source, true, RESPONSE_GIVE, effect, duration);
-    }
-
-    private static int computeDurationInTicks(@Nullable Integer seconds, MobEffect effect) {
+        int count = 0;
+        int duration;
         if (seconds != null) {
             if (effect.isInstantaneous()) {
-                return seconds;
+                duration = seconds;
+            } else if (seconds == -1) {
+                duration = -1;
             } else {
-                return seconds == -1 ? -1 : seconds * 20;
-            }
-        } else {
-            return effect.isInstantaneous() ? 1 : 600;
-        }
+                duration = seconds * 20;
+            }
+        } else if (effect.isInstantaneous()) {
+            duration = 1;
+        } else {
+            duration = 600;
+        }
+
+        for (Entity entity : entities) {
+            if (entity instanceof LivingEntity livingEntity) {
+                MobEffectInstance instance = new MobEffectInstance(effectHolder, duration, amplifier, false, particles);
+                if (livingEntity.addEffect(instance, source.getEntity(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) {
+                    count++;
+                }
+            }
+        }
+
+        if (count == 0) {
+            throw ERROR_GIVE_FAILED.create();
+        }
+
+        if (entities.size() == 1) {
+            source.sendSuccess(
+                () -> Component.translatable(
+                    "commands.effect.give.success.single", effect.getDisplayName(), entities.iterator().next().getDisplayName(), duration / 20
+                ),
+                true
+            );
+        } else {
+            source.sendSuccess(
+                () -> Component.translatable("commands.effect.give.success.multiple", effect.getDisplayName(), entities.size(), duration / 20), true
+            );
+        }
+
+        return count;
     }
 
     private static int clearEffects(CommandSourceStack source, Collection<? extends Entity> entities) throws CommandSyntaxException {
-        CommandResponseTracker<LivingEntity> tracker = CommandResponseTracker.create();
+        int count = 0;
 
         for (Entity entity : entities) {
-            if (entity instanceof LivingEntity livingEntity) {
-                tracker.track(livingEntity, livingEntity.removeAllEffects());
+            if (entity instanceof LivingEntity livingEntity && livingEntity.removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
+                count++;
             }
         }
 
-        return tracker.sendFeedback(source, true, RESPONSE_CLEAR_ALL);
+        if (count == 0) {
+            throw ERROR_CLEAR_EVERYTHING_FAILED.create();
+        }
+
+        if (entities.size() == 1) {
+            source.sendSuccess(
+                () -> Component.translatable("commands.effect.clear.everything.success.single", entities.iterator().next().getDisplayName()), true
+            );
+        } else {
+            source.sendSuccess(() -> Component.translatable("commands.effect.clear.everything.success.multiple", entities.size()), true);
+        }
+
+        return count;
     }
 
     private static int clearEffect(CommandSourceStack source, Collection<? extends Entity> entities, Holder<MobEffect> effectHolder) throws CommandSyntaxException {
         MobEffect effect = effectHolder.value();
-        CommandResponseTracker<LivingEntity> tracker = CommandResponseTracker.create();
+        int count = 0;
 
         for (Entity entity : entities) {
-            if (entity instanceof LivingEntity livingEntity) {
-                tracker.track(livingEntity, livingEntity.removeEffect(effectHolder));
+            if (entity instanceof LivingEntity livingEntity && livingEntity.removeEffect(effectHolder, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.COMMAND)) { // CraftBukkit
+                count++;
             }
         }
 
-        return tracker.sendFeedback(source, true, RESPONSE_CLEAR_SINGLE, effect);
+        if (count == 0) {
+            throw ERROR_CLEAR_SPECIFIC_FAILED.create();
+        }
+
+        if (entities.size() == 1) {
+            source.sendSuccess(
+                () -> Component.translatable(
+                    "commands.effect.clear.specific.success.single", effect.getDisplayName(), entities.iterator().next().getDisplayName()
+                ),
+                true
+            );
+        } else {
+            source.sendSuccess(() -> Component.translatable("commands.effect.clear.specific.success.multiple", effect.getDisplayName(), entities.size()), true);
+        }
+
+        return count;
     }
 }
