--- a/net/minecraft/server/commands/EffectCommands.java
+++ b/net/minecraft/server/commands/EffectCommands.java
@@ -4,7 +_,6 @@
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.arguments.BoolArgumentType;
 import com.mojang.brigadier.arguments.IntegerArgumentType;
-import com.mojang.brigadier.context.CommandContext;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import java.util.Collection;
@@ -21,6 +_,7 @@
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
 
 public class EffectCommands {
     private static final SimpleCommandExceptionType ERROR_GIVE_FAILED = new SimpleCommandExceptionType(Component.translatable("commands.effect.give.failed"));
@@ -181,6 +_,7 @@
         for (Entity entity : p_250411_) {
             if (entity instanceof LivingEntity) {
                 MobEffectInstance mobeffectinstance = new MobEffectInstance(p_249495_, j, p_251498_, false, p_249944_);
+                ((LivingEntity)entity).addEffectCause(EntityPotionEffectEvent.Cause.COMMAND);
                 if (((LivingEntity)entity).addEffect(mobeffectinstance, p_250553_.getEntity())) {
                     i++;
                 }
@@ -211,8 +_,11 @@
         int i = 0;
 
         for (Entity entity : p_136961_) {
-            if (entity instanceof LivingEntity && ((LivingEntity)entity).removeAllEffects()) {
-                i++;
+            if (entity instanceof LivingEntity) {
+                ((LivingEntity)entity).removeAllEffects$cause(EntityPotionEffectEvent.Cause.COMMAND); // CraftBukkit
+                if (((LivingEntity) entity).removeAllEffects()) {
+                    i++;
+                }
             }
         }
 
@@ -236,8 +_,11 @@
         int i = 0;
 
         for (Entity entity : p_248561_) {
-            if (entity instanceof LivingEntity && ((LivingEntity)entity).removeEffect(p_249198_)) {
-                i++;
+            if (entity instanceof LivingEntity livingEntity) {
+                livingEntity.removeEffectNoUpdateCause(EntityPotionEffectEvent.Cause.COMMAND); // CraftBukkit
+                if (livingEntity.removeEffect(p_249198_)) {
+                    i++;
+                }
             }
         }
 
