--- a/net/minecraft/server/commands/FillBiomeCommand.java
+++ b/net/minecraft/server/commands/FillBiomeCommand.java
@@ -89,7 +_,7 @@
     }
 
     private static BiomeResolver makeResolver(MutableInt count, ChunkAccess chunk, BoundingBox region, Holder<Biome> toFill, Predicate<Holder<Biome>> filter) {
-        return (quartX, quartY, quartZ) -> {
+        return (quartX, quartY, quartZ, var8) -> {
             int blockX = QuartPos.toBlock(quartX);
             int blockY = QuartPos.toBlock(quartY);
             int blockZ = QuartPos.toBlock(quartZ);
@@ -146,7 +_,7 @@
         while (iterator.hasNext()) {
             ChunkAccess chunk = iterator.next();
             int previousChangedCount = changedCount.intValue();
-            chunk.fillBiomesFromNoise(makeResolver(changedCount, chunk, region, biome, filter));
+            chunk.fillBiomesFromNoise(makeResolver(changedCount, chunk, region, biome, filter), level.getChunkSource().randomState().sampler());
             if (previousChangedCount != changedCount.intValue()) {
                 chunk.markUnsaved();
             } else {
