--- a/net/minecraft/server/commands/ForceLoadCommand.java
+++ b/net/minecraft/server/commands/ForceLoadCommand.java
@@ -30,32 +_,6 @@
     private static final SimpleCommandExceptionType ERROR_NONE_REMOVED = new SimpleCommandExceptionType(
         Component.translatable("commands.forceload.removed.failure")
     );
-    private static final CommandResponseTracker.MessagesWithArg<ChunkPos, ForceLoadCommand.ForceLoadParams> RESPONSE_ADD = CommandResponseTracker.messages(
-        ERROR_ALL_ADDED,
-        (chunkPos, var1, params) -> Component.translatable(
-            "commands.forceload.added.single", Component.translationArg(chunkPos), Component.translationArg(params.dimension().identifier())
-        ),
-        (chunkCount, var1, params) -> Component.translatable(
-            "commands.forceload.added.multiple",
-            chunkCount,
-            Component.translationArg(params.dimension().identifier()),
-            Component.translationArg(params.min()),
-            Component.translationArg(params.max())
-        )
-    );
-    private static final CommandResponseTracker.MessagesWithArg<ChunkPos, ForceLoadCommand.ForceLoadParams> RESPONSE_REMOVE = CommandResponseTracker.messages(
-        ERROR_NONE_REMOVED,
-        (chunkPos, var1, params) -> Component.translatable(
-            "commands.forceload.removed.single", Component.translationArg(chunkPos), Component.translationArg(params.dimension().identifier())
-        ),
-        (chunkCount, var1, params) -> Component.translatable(
-            "commands.forceload.removed.multiple",
-            chunkCount,
-            Component.translationArg(params.dimension().identifier()),
-            Component.translationArg(params.min()),
-            Component.translationArg(params.max())
-        )
-    );
 
     public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
         dispatcher.register(
@@ -119,13 +_,13 @@
         if (result) {
             source.sendSuccess(
                 () -> Component.translatable(
-                    "commands.forceload.query.success", Component.translationArg(chunkPos), Component.translationArg(dimension.identifier())
+                        "commands.forceload.query.success", Component.translationArg(chunkPos), level.getDescription() // Neo: Use dimension translation, if one exists
                 ),
                 false
             );
             return 1;
         } else {
-            throw ERROR_NOT_TICKING.create(chunkPos, dimension.identifier());
+            throw ERROR_NOT_TICKING.create(chunkPos, level.getDescription()); // Neo: Use dimension translation, if one exists
         }
     }
 
@@ -135,19 +_,19 @@
         LongSet forcedChunks = level.getForceLoadedChunks();
         int chunkCount = forcedChunks.size();
         if (chunkCount > 0) {
-            String chunkList = Joiner.on(", ").join(forcedChunks.longStream().sorted().mapToObj(ChunkPos::unpack).map(ChunkPos::toString).iterator());
+            String chunkList = Joiner.on(", ").join(forcedChunks.stream().sorted().map(ChunkPos::unpack).map(ChunkPos::toString).iterator());
             if (chunkCount == 1) {
                 source.sendSuccess(
-                    () -> Component.translatable("commands.forceload.list.single", Component.translationArg(dimension.identifier()), chunkList), false
+                    () -> Component.translatable("commands.forceload.list.single", level.getDescription(), chunkList), false // Neo: Use dimension translation, if one exists
                 );
             } else {
                 source.sendSuccess(
-                    () -> Component.translatable("commands.forceload.list.multiple", chunkCount, Component.translationArg(dimension.identifier()), chunkList),
+                    () -> Component.translatable("commands.forceload.list.multiple", chunkCount, level.getDescription(), chunkList), // Neo: Use dimension translation, if one exists
                     false
                 );
             }
         } else {
-            source.sendFailure(Component.translatable("commands.forceload.added.none", Component.translationArg(dimension.identifier())));
+            source.sendFailure(Component.translatable("commands.forceload.added.none", level.getDescription())); // Neo: Use dimension translation, if one exists
         }
 
         return chunkCount;
@@ -158,7 +_,7 @@
         ResourceKey<Level> dimension = level.dimension();
         LongSet forcedChunks = level.getForceLoadedChunks();
         forcedChunks.forEach(chunk -> level.setChunkForced(ChunkPos.getX(chunk), ChunkPos.getZ(chunk), false));
-        source.sendSuccess(() -> Component.translatable("commands.forceload.removed.all", Component.translationArg(dimension.identifier())), true);
+        source.sendSuccess(() -> Component.translatable("commands.forceload.removed.all", level.getDescription()), true); // Neo: Use dimension translation, if one exists
         return 0;
     }
 
@@ -177,31 +_,56 @@
                 throw ERROR_TOO_MANY_CHUNKS.create(256, chunkCount);
             }
 
-            CommandResponseTracker<ChunkPos> tracker = CommandResponseTracker.create();
             ServerLevel level = source.getLevel();
             ResourceKey<Level> dimension = level.dimension();
+            ChunkPos firstChanged = null;
+            int changedCount = 0;
 
             for (int x = minChunkX; x <= maxChunkX; x++) {
                 for (int z = minChunkZ; z <= maxChunkZ; z++) {
-                    tracker.track(new ChunkPos(x, z), level.setChunkForced(x, z, add));
+                    boolean changed = level.setChunkForced(x, z, add);
+                    if (changed) {
+                        changedCount++;
+                        if (firstChanged == null) {
+                            firstChanged = new ChunkPos(x, z);
+                        }
+                    }
                 }
             }
 
-            return tracker.sendFeedback(
-                source, true, add ? RESPONSE_ADD : RESPONSE_REMOVE, new ForceLoadCommand.ForceLoadParams(dimension, minChunkX, minChunkZ, maxChunkX, maxChunkZ)
-            );
+            ChunkPos finalFirstChanged = firstChanged;
+            int changedChunks = changedCount;
+            if (changedChunks == 0) {
+                throw (add ? ERROR_ALL_ADDED : ERROR_NONE_REMOVED).create();
+            }
+
+            if (changedChunks == 1) {
+                source.sendSuccess(
+                    () -> Component.translatable(
+                        "commands.forceload." + (add ? "added" : "removed") + ".single",
+                        Component.translationArg(finalFirstChanged),
+                        level.getDescription() // Neo: Use dimension translation, if one exists
+                    ),
+                    true
+                );
+            } else {
+                ChunkPos min = new ChunkPos(minChunkX, minChunkZ);
+                ChunkPos max = new ChunkPos(maxChunkX, maxChunkZ);
+                source.sendSuccess(
+                    () -> Component.translatable(
+                        "commands.forceload." + (add ? "added" : "removed") + ".multiple",
+                        changedChunks,
+                        level.getDescription(), // Neo: Use dimension translation, if one exists
+                        Component.translationArg(min),
+                        Component.translationArg(max)
+                    ),
+                    true
+                );
+            }
+
+            return changedChunks;
         } else {
             throw BlockPosArgument.ERROR_OUT_OF_WORLD.create();
-        }
-    }
-
-    private record ForceLoadParams(ResourceKey<Level> dimension, int minChunkX, int minChunkZ, int maxChunkX, int maxChunkZ) {
-        public ChunkPos min() {
-            return new ChunkPos(this.minChunkX, this.minChunkZ);
-        }
-
-        public ChunkPos max() {
-            return new ChunkPos(this.maxChunkX, this.maxChunkZ);
         }
     }
 }
