--- a/net/minecraft/server/commands/GameModeCommand.java
+++ b/net/minecraft/server/commands/GameModeCommand.java
@@ -9,6 +_,7 @@
 import net.minecraft.commands.arguments.EntityArgument;
 import net.minecraft.commands.arguments.GameModeArgument;
 import net.minecraft.network.chat.Component;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.permissions.PermissionCheck;
 import net.minecraft.server.permissions.Permissions;
@@ -47,9 +_,26 @@
     }
 
     private static int setMode(CommandContext<CommandSourceStack> context, Collection<ServerPlayer> players, GameType type) {
+        // Purpur start - Gamemode extra permissions
+        if (org.purpurmc.purpur.PurpurConfig.commandGamemodeRequiresPermission) {
+            String gamemode = type.getName();
+            CommandSourceStack sender = context.getSource();
+            if (!sender.testPermission(Permissions.COMMANDS_GAMEMASTER, "minecraft.command.gamemode." + gamemode)) {
+                return 0;
+            }
+            if (sender.getEntity() instanceof ServerPlayer player && (players.size() > 1 || !players.contains(player)) && !sender.testPermission(Permissions.COMMANDS_GAMEMASTER, "minecraft.command.gamemode." + gamemode + ".other")) {
+                return 0;
+            }
+        }
+        // Purpur end - Gamemode extra permissions
         int count = 0;
+        MinecraftServer server = context.getSource().getServer();
 
         for (ServerPlayer player : players) {
+            if (server.isSingleplayerOwner(player.nameAndId())) {
+                server.setDefaultGameType(type);
+            }
+
             if (setGameMode(context.getSource(), player, type)) {
                 count++;
             }
@@ -62,10 +_,27 @@
         setGameMode(player.createCommandSourceStack(), player, type);
     }
 
-    private static boolean setGameMode(CommandSourceStack source, ServerPlayer player, GameType type) {
-        if (player.setGameMode(type)) {
+    private static boolean setGameMode(final CommandSourceStack source, final ServerPlayer player, final GameType type) {
+        // Paper start - Expand PlayerGameModeChangeEvent
+        return setGameMode(source, player, type, org.bukkit.event.player.PlayerGameModeChangeEvent.Cause.COMMAND);
+    }
+
+    public static boolean setGameMode(final CommandSourceStack source, final ServerPlayer player, final GameType type, final org.bukkit.event.player.PlayerGameModeChangeEvent.Cause cause) {
+        org.bukkit.event.player.PlayerGameModeChangeEvent event = player.setGameMode(type, cause, null);
+        if (event != null && !event.isCancelled()) {
+            // Paper end - Expand PlayerGameModeChangeEvent
+            MinecraftServer server = source.getServer();
+            if (server.isSingleplayerOwner(player.nameAndId())) {
+                server.setDefaultGameType(type);
+            }
+
             logGamemodeChange(source, player, type);
             return true;
+            // Paper start - Expand PlayerGameModeChangeEvent
+        } else if (event != null && event.cancelMessage() != null) {
+            source.sendSuccess(() -> io.papermc.paper.adventure.PaperAdventure.asVanilla(event.cancelMessage()), true);
+            return false;
+            // Paper end - Expand PlayerGameModeChangeEvent
         } else {
             return false;
         }
