--- a/net/minecraft/server/commands/GameRuleCommand.java
+++ b/net/minecraft/server/commands/GameRuleCommand.java
@@ -47,7 +_,7 @@
             throw ERROR_GAME_RULE_NOT_SET.create(gameRule, seralizedValue);
         }
 
-        gameRules.set(gameRule, value, context.getSource().getServer());
+        // gameRules.set(gameRule, value, context.getSource().getServer()); // Paper - per-world game rules and event - handle in CraftEventFactory.handleGameRuleSet
         source.sendSuccess(() -> Component.translatable("commands.gamerule.set", gameRule.id(), seralizedValue), true);
         return gameRule.getCommandResult(value);
     }
