--- a/net/minecraft/server/commands/GiveCommand.java
+++ b/net/minecraft/server/commands/GiveCommand.java
@@ -2,7 +_,6 @@
 
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.arguments.IntegerArgumentType;
-import com.mojang.brigadier.context.CommandContext;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import java.util.Collection;
 import net.minecraft.commands.CommandBuildContext;
@@ -67,7 +_,7 @@
                     ItemStack itemstack1 = p_137780_.createItemStack(l, false);
                     boolean flag = serverplayer.getInventory().add(itemstack1);
                     if (flag && itemstack1.isEmpty()) {
-                        ItemEntity itementity1 = serverplayer.drop(itemstack, false);
+                        ItemEntity itementity1 = serverplayer.drop(itemstack, false, false, false);  // CraftBukkit - SPIGOT-2942: Add boolean to call event
                         if (itementity1 != null) {
                             itementity1.makeFakeItem();
                         }
