--- a/net/minecraft/server/commands/KickCommand.java
+++ b/net/minecraft/server/commands/KickCommand.java
@@ -1,7 +_,6 @@
 package net.minecraft.server.commands;
 
 import com.mojang.brigadier.CommandDispatcher;
-import com.mojang.brigadier.context.CommandContext;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import java.util.Collection;
@@ -53,7 +_,7 @@
 
             for (ServerPlayer serverplayer : p_137803_) {
                 if (!p_137802_.getServer().isSingleplayerOwner(serverplayer.getGameProfile())) {
-                    serverplayer.connection.disconnect(p_137804_);
+                    serverplayer.connection.disconnect(p_137804_, org.bukkit.event.player.PlayerKickEvent.Cause.KICK_COMMAND); // Paper - kick event cause
                     p_137802_.sendSuccess(() -> Component.translatable("commands.kick.success", serverplayer.getDisplayName(), p_137804_), true);
                     i++;
                 }
