--- a/net/minecraft/server/commands/LootCommand.java
+++ b/net/minecraft/server/commands/LootCommand.java
@@ -1,5 +_,6 @@
 package net.minecraft.server.commands;
 
+import com.google.common.collect.Lists;
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.arguments.IntegerArgumentType;
 import com.mojang.brigadier.builder.ArgumentBuilder;
@@ -22,8 +_,6 @@
 import net.minecraft.core.Holder;
 import net.minecraft.network.chat.Component;
 import net.minecraft.resources.ResourceKey;
-import net.minecraft.server.commands.item.BlockItemAccessor;
-import net.minecraft.server.commands.item.ItemCommands;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.Container;
@@ -53,16 +_,6 @@
     private static final DynamicCommandExceptionType ERROR_NO_BLOCK_LOOT_TABLE = new DynamicCommandExceptionType(
         block -> Component.translatableEscape("commands.drop.no_loot_table.block", block)
     );
-    private static final CommandResponseTracker.Messages<ItemStack> RESPONSE_WITHOUT_LOOT_TABLE = CommandResponseTracker.messages(
-        (drop, var1) -> Component.translatable("commands.drop.success.single", drop.getCount(), drop.getDisplayName()),
-        (dropCount, var1) -> Component.translatable("commands.drop.success.multiple", dropCount)
-    );
-    private static final CommandResponseTracker.MessagesWithArg<ItemStack, ResourceKey<LootTable>> RESPONSE_WITH_LOOT_TABLE = CommandResponseTracker.messages(
-        (drop, var1, key) -> Component.translatable(
-            "commands.drop.success.single_with_table", drop.getCount(), drop.getDisplayName(), Component.translationArg(key.identifier())
-        ),
-        (dropCount, var1, key) -> Component.translatable("commands.drop.success.multiple_with_table", dropCount, Component.translationArg(key.identifier()))
-    );
 
     public static void register(CommandDispatcher<CommandSourceStack> dispatcher, CommandBuildContext context) {
         dispatcher.register(
@@ -187,19 +_,19 @@
                                     .then(
                                         tail.construct(
                                                 Commands.argument("slot", SlotArgument.slot()),
-                                                (c, drops, usedItems) -> entityReplace(
-                                                    EntityArgument.getEntities(c, "entities"), SlotArgument.getSlot(c, "slot"), drops.size(), drops, usedItems
+                                                (c, drops, callback) -> entityReplace(
+                                                    EntityArgument.getEntities(c, "entities"), SlotArgument.getSlot(c, "slot"), drops.size(), drops, callback
                                                 )
                                             )
                                             .then(
                                                 tail.construct(
                                                     Commands.argument("count", IntegerArgumentType.integer(0)),
-                                                    (c, drops, usedItems) -> entityReplace(
+                                                    (c, drops, callback) -> entityReplace(
                                                         EntityArgument.getEntities(c, "entities"),
                                                         SlotArgument.getSlot(c, "slot"),
                                                         IntegerArgumentType.getInteger(c, "count"),
                                                         drops,
-                                                        usedItems
+                                                        callback
                                                     )
                                                 )
                                             )
@@ -213,25 +_,25 @@
                                     .then(
                                         tail.construct(
                                                 Commands.argument("slot", SlotArgument.slot()),
-                                                (c, drops, usedItems) -> blockReplace(
+                                                (c, drops, callback) -> blockReplace(
                                                     c.getSource(),
                                                     BlockPosArgument.getLoadedBlockPos(c, "targetPos"),
                                                     SlotArgument.getSlot(c, "slot"),
                                                     drops.size(),
                                                     drops,
-                                                    usedItems
+                                                    callback
                                                 )
                                             )
                                             .then(
                                                 tail.construct(
                                                     Commands.argument("count", IntegerArgumentType.integer(0)),
-                                                    (c, drops, usedItems) -> blockReplace(
+                                                    (c, drops, callback) -> blockReplace(
                                                         c.getSource(),
                                                         BlockPosArgument.getLoadedBlockPos(c, "targetPos"),
                                                         IntegerArgumentType.getInteger(c, "slot"),
                                                         IntegerArgumentType.getInteger(c, "count"),
                                                         drops,
-                                                        usedItems
+                                                        callback
                                                     )
                                                 )
                                             )
@@ -244,7 +_,7 @@
                     .then(
                         tail.construct(
                             Commands.argument("targetPos", BlockPosArgument.blockPos()),
-                            (c, drops, usedItems) -> blockDistribute(c.getSource(), BlockPosArgument.getLoadedBlockPos(c, "targetPos"), drops, usedItems)
+                            (c, drops, callback) -> blockDistribute(c.getSource(), BlockPosArgument.getLoadedBlockPos(c, "targetPos"), drops, callback)
                         )
                     )
             )
@@ -253,7 +_,7 @@
                     .then(
                         tail.construct(
                             Commands.argument("players", EntityArgument.players()),
-                            (c, drops, usedItems) -> playerGive(EntityArgument.getPlayers(c, "players"), drops, usedItems)
+                            (c, drops, callback) -> playerGive(EntityArgument.getPlayers(c, "players"), drops, callback)
                         )
                     )
             )
@@ -262,21 +_,33 @@
                     .then(
                         tail.construct(
                             Commands.argument("targetPos", Vec3Argument.vec3()),
-                            (c, drops, usedItems) -> dropInWorld(c.getSource(), Vec3Argument.getVec3(c, "targetPos"), drops, usedItems)
+                            (c, drops, callback) -> dropInWorld(c.getSource(), Vec3Argument.getVec3(c, "targetPos"), drops, callback)
                         )
                     )
             );
     }
 
-    private static void blockDistribute(CommandSourceStack source, BlockPos pos, List<ItemStack> drops, CommandResponseTracker<ItemStack> usedItems) throws CommandSyntaxException {
-        Container container = BlockItemAccessor.getContainer(source, pos, ItemCommands.ERROR_TARGET_NOT_A_CONTAINER);
+    private static Container getContainer(CommandSourceStack source, BlockPos pos) throws CommandSyntaxException {
+        if (source.getLevel().getBlockEntity(pos) instanceof Container container) {
+            return container;
+        } else {
+            throw ItemCommands.ERROR_TARGET_NOT_A_CONTAINER.create(pos.getX(), pos.getY(), pos.getZ());
+        }
+    }
+
+    private static int blockDistribute(CommandSourceStack source, BlockPos pos, List<ItemStack> drops, LootCommand.Callback callback) throws CommandSyntaxException {
+        Container container = getContainer(source, pos);
+        List<ItemStack> usedItems = Lists.newArrayListWithCapacity(drops.size());
 
         for (ItemStack drop : drops) {
             if (distributeToContainer(container, drop.copy())) {
                 container.setChanged();
-                usedItems.track(drop);
+                usedItems.add(drop);
             }
         }
+
+        callback.accept(usedItems);
+        return usedItems.size();
     }
 
     private static boolean distributeToContainer(Container container, ItemStack itemStack) {
@@ -304,20 +_,23 @@
         return changed;
     }
 
-    private static void blockReplace(
-        CommandSourceStack source, BlockPos pos, int startSlot, int slotCount, List<ItemStack> drops, CommandResponseTracker<ItemStack> usedItems
-    ) throws CommandSyntaxException {
-        Container container = BlockItemAccessor.getContainer(source, pos, ItemCommands.ERROR_TARGET_NOT_A_CONTAINER);
+    private static int blockReplace(CommandSourceStack source, BlockPos pos, int startSlot, int slotCount, List<ItemStack> drops, LootCommand.Callback callback) throws CommandSyntaxException {
+        Container container = getContainer(source, pos);
         int maxSlot = container.getContainerSize();
         if (startSlot >= 0 && startSlot < maxSlot) {
+            List<ItemStack> usedItems = Lists.newArrayListWithCapacity(drops.size());
+
             for (int i = 0; i < slotCount; i++) {
                 int slot = startSlot + i;
                 ItemStack toAdd = i < drops.size() ? drops.get(i) : ItemStack.EMPTY;
                 if (container.canPlaceItem(slot, toAdd)) {
                     container.setItem(slot, toAdd);
-                    usedItems.track(toAdd);
+                    usedItems.add(toAdd);
                 }
             }
+
+            callback.accept(usedItems);
+            return usedItems.size();
         } else {
             throw ItemCommands.ERROR_TARGET_INAPPLICABLE_SLOT.create(startSlot);
         }
@@ -327,29 +_,34 @@
         return a.getCount() <= a.getMaxStackSize() && ItemStack.isSameItemSameComponents(a, b);
     }
 
-    private static void playerGive(Collection<ServerPlayer> players, List<ItemStack> drops, CommandResponseTracker<ItemStack> usedItems) {
+    private static int playerGive(Collection<ServerPlayer> players, List<ItemStack> drops, LootCommand.Callback callback) throws CommandSyntaxException {
+        List<ItemStack> usedItems = Lists.newArrayListWithCapacity(drops.size());
+
         for (ItemStack drop : drops) {
             for (ServerPlayer player : players) {
                 if (player.getInventory().add(drop.copy())) {
-                    usedItems.track(drop);
+                    usedItems.add(drop);
                 }
             }
         }
+
+        callback.accept(usedItems);
+        return usedItems.size();
     }
 
-    private static void setSlots(Entity entity, List<ItemStack> itemsToSet, int startSlot, int count, CommandResponseTracker<ItemStack> usedItems) {
+    private static void setSlots(Entity entity, List<ItemStack> itemsToSet, int startSlot, int count, List<ItemStack> usedItems) {
         for (int i = 0; i < count; i++) {
             ItemStack item = i < itemsToSet.size() ? itemsToSet.get(i) : ItemStack.EMPTY;
             SlotAccess slotAccess = entity.getSlot(startSlot + i);
             if (slotAccess != null && slotAccess.set(item.copy())) {
-                usedItems.track(item);
+                usedItems.add(item);
             }
         }
     }
 
-    private static void entityReplace(
-        Collection<? extends Entity> entities, int startSlot, int count, List<ItemStack> drops, CommandResponseTracker<ItemStack> usedItems
-    ) {
+    private static int entityReplace(Collection<? extends Entity> entities, int startSlot, int count, List<ItemStack> drops, LootCommand.Callback callback) throws CommandSyntaxException {
+        List<ItemStack> usedItems = Lists.newArrayListWithCapacity(drops.size());
+
         for (Entity entity : entities) {
             if (entity instanceof ServerPlayer player) {
                 setSlots(entity, drops, startSlot, count, usedItems);
@@ -358,16 +_,46 @@
                 setSlots(entity, drops, startSlot, count, usedItems);
             }
         }
+
+        callback.accept(usedItems);
+        return usedItems.size();
     }
 
-    private static void dropInWorld(CommandSourceStack source, Vec3 pos, List<ItemStack> drops, CommandResponseTracker<ItemStack> usedItems) {
+    private static int dropInWorld(CommandSourceStack source, Vec3 pos, List<ItemStack> drops, LootCommand.Callback callback) throws CommandSyntaxException {
         ServerLevel level = source.getLevel();
+        drops.removeIf(ItemStack::isEmpty); // CraftBukkit - SPIGOT-6959 Remove empty items for avoid throw an error in new EntityItem
         drops.forEach(drop -> {
             ItemEntity entity = new ItemEntity(level, pos.x, pos.y, pos.z, drop.copy());
             entity.setDefaultPickUpDelay();
             level.addFreshEntity(entity);
-            usedItems.track(drop);
         });
+        callback.accept(drops);
+        return drops.size();
+    }
+
+    private static void callback(CommandSourceStack source, List<ItemStack> drops) {
+        if (drops.size() == 1) {
+            ItemStack drop = drops.get(0);
+            source.sendSuccess(() -> Component.translatable("commands.drop.success.single", drop.getCount(), drop.getDisplayName()), false);
+        } else {
+            source.sendSuccess(() -> Component.translatable("commands.drop.success.multiple", drops.size()), false);
+        }
+    }
+
+    private static void callback(CommandSourceStack source, List<ItemStack> drops, ResourceKey<LootTable> location) {
+        if (drops.size() == 1) {
+            ItemStack drop = drops.get(0);
+            source.sendSuccess(
+                () -> Component.translatable(
+                    "commands.drop.success.single_with_table", drop.getCount(), drop.getDisplayName(), Component.translationArg(location.identifier())
+                ),
+                false
+            );
+        } else {
+            source.sendSuccess(
+                () -> Component.translatable("commands.drop.success.multiple_with_table", drops.size(), Component.translationArg(location.identifier())), false
+            );
+        }
     }
 
     private static ItemStack getSourceHandItem(CommandSourceStack source, EquipmentSlot slot) throws CommandSyntaxException {
@@ -396,9 +_,7 @@
             .withOptionalParameter(LootContextParams.THIS_ENTITY, source.getEntity())
             .withParameter(LootContextParams.TOOL, tool);
         List<ItemStack> drops = blockState.getDrops(lootParams);
-        CommandResponseTracker<ItemStack> usedItems = CommandResponseTracker.create();
-        output.accept(context, drops, usedItems);
-        return usedItems.sendFeedback(source, false, RESPONSE_WITH_LOOT_TABLE, lootTable.get());
+        return output.accept(context, drops, usedItems -> callback(source, usedItems, lootTable.get()));
     }
 
     private static int dropKillLoot(CommandContext<CommandSourceStack> context, Entity target, LootCommand.DropConsumer output) throws CommandSyntaxException {
@@ -422,9 +_,7 @@
         LootParams lootParams = builder.create(LootContextParamSets.ENTITY);
         LootTable lootTable = source.getServer().reloadableRegistries().getLootTable(lootTableId.get());
         List<ItemStack> drops = lootTable.getRandomItems(lootParams);
-        CommandResponseTracker<ItemStack> usedItems = CommandResponseTracker.create();
-        output.accept(context, drops, usedItems);
-        return usedItems.sendFeedback(source, false, RESPONSE_WITH_LOOT_TABLE, lootTableId.get());
+        return output.accept(context, drops, usedItems -> callback(source, usedItems, lootTableId.get()));
     }
 
     private static int dropChestLoot(CommandContext<CommandSourceStack> context, Holder<LootTable> lootTable, LootCommand.DropConsumer output) throws CommandSyntaxException {
@@ -451,14 +_,17 @@
     private static int drop(CommandContext<CommandSourceStack> context, Holder<LootTable> lootTable, LootParams lootParams, LootCommand.DropConsumer output) throws CommandSyntaxException {
         CommandSourceStack source = context.getSource();
         List<ItemStack> drops = lootTable.value().getRandomItems(lootParams);
-        CommandResponseTracker<ItemStack> usedItems = CommandResponseTracker.create();
-        output.accept(context, drops, usedItems);
-        return usedItems.sendFeedback(source, false, RESPONSE_WITHOUT_LOOT_TABLE);
+        return output.accept(context, drops, usedItems -> callback(source, usedItems));
+    }
+
+    @FunctionalInterface
+    private interface Callback {
+        void accept(List<ItemStack> setItems) throws CommandSyntaxException;
     }
 
     @FunctionalInterface
     private interface DropConsumer {
-        void accept(CommandContext<CommandSourceStack> context, List<ItemStack> drops, CommandResponseTracker<ItemStack> response) throws CommandSyntaxException;
+        int accept(CommandContext<CommandSourceStack> context, List<ItemStack> drops, LootCommand.Callback successCallback) throws CommandSyntaxException;
     }
 
     @FunctionalInterface
