--- a/net/minecraft/server/commands/OpCommand.java
+++ b/net/minecraft/server/commands/OpCommand.java
@@ -2,17 +_,14 @@
 
 import com.mojang.authlib.GameProfile;
 import com.mojang.brigadier.CommandDispatcher;
-import com.mojang.brigadier.context.CommandContext;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
-import com.mojang.brigadier.suggestion.SuggestionsBuilder;
 import java.util.Collection;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.SharedSuggestionProvider;
 import net.minecraft.commands.arguments.GameProfileArgument;
 import net.minecraft.network.chat.Component;
-import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.server.players.PlayerList;
 
 public class OpCommand {
@@ -49,7 +_,7 @@
             if (!playerlist.isOp(gameprofile)) {
                 playerlist.op(gameprofile);
                 i++;
-                p_138089_.sendSuccess(() -> Component.translatable("commands.op.success", p_138090_.iterator().next().getName()), true);
+                p_138089_.sendSuccess(() -> Component.translatable("commands.op.success", gameprofile.getName()), true); // Paper - fixes MC-253721
             }
         }
 
