--- a/net/minecraft/server/commands/PardonCommand.java
+++ b/net/minecraft/server/commands/PardonCommand.java
@@ -27,7 +_,7 @@
         );
     }
 
-    private static int pardonPlayers(CommandSourceStack source, Collection<NameAndId> players) throws CommandSyntaxException {
+    public static int pardonPlayers(CommandSourceStack source, Collection<NameAndId> players) throws CommandSyntaxException {
         UserBanList list = source.getServer().getPlayerList().getBans();
         int count = 0;
 
