--- a/net/minecraft/server/commands/PardonIpCommand.java
+++ b/net/minecraft/server/commands/PardonIpCommand.java
@@ -33,7 +_,7 @@
         );
     }
 
-    private static int unban(CommandSourceStack p_138118_, String p_138119_) throws CommandSyntaxException {
+    public static int unban(CommandSourceStack p_138118_, String p_138119_) throws CommandSyntaxException {
         if (!InetAddresses.isInetAddress(p_138119_)) {
             throw ERROR_INVALID.create();
         } else {
