--- a/net/minecraft/server/commands/PardonIpCommand.java
+++ b/net/minecraft/server/commands/PardonIpCommand.java
@@ -27,7 +_,7 @@
         );
     }
 
-    private static int unban(CommandSourceStack source, String ip) throws CommandSyntaxException {
+    public static int unban(CommandSourceStack source, String ip) throws CommandSyntaxException {
         if (!InetAddresses.isInetAddress(ip)) {
             throw ERROR_INVALID.create();
         } else {
