--- a/net/minecraft/server/commands/PardonCommand.java
+++ b/net/minecraft/server/commands/PardonCommand.java
@@ -33,7 +_,7 @@
         );
     }
 
-    private static int pardonPlayers(CommandSourceStack p_138103_, Collection<GameProfile> p_138104_) throws CommandSyntaxException {
+    public static int pardonPlayers(CommandSourceStack p_138103_, Collection<GameProfile> p_138104_) throws CommandSyntaxException {
         UserBanList userbanlist = p_138103_.getServer().getPlayerList().getBans();
         int i = 0;
 
