--- a/net/minecraft/server/commands/PlaceCommand.java
+++ b/net/minecraft/server/commands/PlaceCommand.java
@@ -3,20 +_,17 @@
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.arguments.FloatArgumentType;
 import com.mojang.brigadier.arguments.IntegerArgumentType;
-import com.mojang.brigadier.builder.RequiredArgumentBuilder;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import com.mojang.brigadier.suggestion.SuggestionProvider;
 import java.util.Optional;
 import net.minecraft.IdentifierException;
-import net.minecraft.commands.CommandBuildContext;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.SharedSuggestionProvider;
 import net.minecraft.commands.arguments.IdentifierArgument;
 import net.minecraft.commands.arguments.ResourceKeyArgument;
-import net.minecraft.commands.arguments.ResourceOrIdArgument;
 import net.minecraft.commands.arguments.TemplateMirrorArgument;
 import net.minecraft.commands.arguments.TemplateRotationArgument;
 import net.minecraft.commands.arguments.coordinates.BlockPosArgument;
@@ -32,8 +_,6 @@
 import net.minecraft.world.level.block.Rotation;
 import net.minecraft.world.level.block.entity.StructureBlockEntity;
 import net.minecraft.world.level.chunk.ChunkGenerator;
-import net.minecraft.world.level.levelgen.RandomState;
-import net.minecraft.world.level.levelgen.densityfunction.SamplerContext;
 import net.minecraft.world.level.levelgen.feature.Feature;
 import net.minecraft.world.level.levelgen.structure.BoundingBox;
 import net.minecraft.world.level.levelgen.structure.Structure;
@@ -60,30 +_,30 @@
         Component.translatable("commands.place.template.failed")
     );
     private static final SuggestionProvider<CommandSourceStack> SUGGEST_TEMPLATES = (context, builder) -> {
-        StructureTemplateManager structureManager = context.getSource().getLevel().getStructureTemplateManager();
+        StructureTemplateManager structureManager = context.getSource().getLevel().getStructureManager();
         return SharedSuggestionProvider.suggestResource(structureManager.listTemplates(), builder);
     };
 
-    public static void register(CommandDispatcher<CommandSourceStack> dispatcher, CommandBuildContext context) {
+    public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
         dispatcher.register(
             Commands.literal("place")
                 .requires(Commands.hasPermission(Commands.LEVEL_GAMEMASTERS))
                 .then(
                     Commands.literal("feature")
                         .then(
-                            (Commands.argument("feature", ResourceOrIdArgument.feature(context))
-                                    .executes(
-                                        c -> placeFeature(
-                                            (CommandSourceStack)c.getSource(),
-                                            ResourceOrIdArgument.getFeature(c, "feature"),
-                                            BlockPos.containing(((CommandSourceStack)c.getSource()).getPosition())
-                                        )
-                                    ))
+                            Commands.argument("feature", ResourceKeyArgument.key(Registries.FEATURE))
+                                .executes(
+                                    c -> placeFeature(
+                                        c.getSource(), ResourceKeyArgument.getConfiguredFeature(c, "feature"), BlockPos.containing(c.getSource().getPosition())
+                                    )
+                                )
                                 .then(
                                     Commands.argument("pos", BlockPosArgument.blockPos())
                                         .executes(
                                             c -> placeFeature(
-                                                c.getSource(), ResourceOrIdArgument.getFeature(c, "feature"), BlockPosArgument.getLoadedBlockPos(c, "pos")
+                                                c.getSource(),
+                                                ResourceKeyArgument.getConfiguredFeature(c, "feature"),
+                                                BlockPosArgument.getLoadedBlockPos(c, "pos")
                                             )
                                         )
                                 )
@@ -253,7 +_,7 @@
         );
     }
 
-    public static int placeFeature(CommandSourceStack source, Holder<Feature> featureHolder, BlockPos pos) throws CommandSyntaxException {
+    public static int placeFeature(CommandSourceStack source, Holder.Reference<Feature> featureHolder, BlockPos pos) throws CommandSyntaxException {
         ServerLevel level = source.getLevel();
         Feature feature = featureHolder.value();
         ChunkPos chunkPos = ChunkPos.containing(pos);
@@ -262,14 +_,8 @@
             throw ERROR_FEATURE_FAILED.create();
         }
 
-        Optional<String> key = featureHolder.getRegisteredNameIfPresent();
-        if (key.isPresent()) {
-            String id = key.get();
-            source.sendSuccess(() -> Component.translatable("commands.place.feature.success", id, pos.getX(), pos.getY(), pos.getZ()), true);
-        } else {
-            source.sendSuccess(() -> Component.translatable("commands.place.feature.success.inline", pos.getX(), pos.getY(), pos.getZ()), true);
-        }
-
+        String id = featureHolder.key().identifier().toString();
+        source.sendSuccess(() -> Component.translatable("commands.place.feature.success", id, pos.getX(), pos.getY(), pos.getZ()), true);
         return 1;
     }
 
@@ -289,16 +_,14 @@
         ServerLevel level = source.getLevel();
         Structure structure = structureHolder.value();
         ChunkGenerator chunkGenerator = level.getChunkSource().getGenerator();
-        RandomState randomState = level.getChunkSource().randomState();
         StructureStart start = structure.generate(
             structureHolder,
             level.dimension(),
             source.registryAccess(),
             chunkGenerator,
             chunkGenerator.getBiomeSource(),
-            randomState.createClimateSampler(SamplerContext.EMPTY_UNCACHED),
-            randomState,
-            level.getStructureTemplateManager(),
+            level.getChunkSource().randomState(),
+            level.getStructureManager(),
             level.getSeed(),
             ChunkPos.containing(pos),
             0,
@@ -309,6 +_,7 @@
             throw ERROR_STRUCTURE_FAILED.create();
         }
 
+        start.generationEventCause = org.bukkit.event.world.AsyncStructureGenerateEvent.Cause.COMMAND; // CraftBukkit - set AsyncStructureGenerateEvent.Cause.COMMAND as generation cause
         BoundingBox boundingBox = start.getBoundingBox();
         ChunkPos chunkMin = new ChunkPos(SectionPos.blockToSectionCoord(boundingBox.minX()), SectionPos.blockToSectionCoord(boundingBox.minZ()));
         ChunkPos chunkMax = new ChunkPos(SectionPos.blockToSectionCoord(boundingBox.maxX()), SectionPos.blockToSectionCoord(boundingBox.maxZ()));
@@ -333,7 +_,7 @@
         CommandSourceStack source, Identifier template, BlockPos pos, Rotation rotation, Mirror mirror, float integrity, int seed, boolean strict
     ) throws CommandSyntaxException {
         ServerLevel level = source.getLevel();
-        StructureTemplateManager manager = level.getStructureTemplateManager();
+        StructureTemplateManager manager = level.getStructureManager();
 
         Optional<StructureTemplate> maybeStructureTemplate;
         try {
