--- a/net/minecraft/server/commands/PublishCommand.java
+++ b/net/minecraft/server/commands/PublishCommand.java
@@ -8,11 +_,14 @@
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
+import net.minecraft.commands.arguments.GameModeArgument;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.ComponentUtils;
 import net.minecraft.network.chat.MutableComponent;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.HttpUtil;
+import net.minecraft.world.level.GameType;
+import org.jspecify.annotations.Nullable;
 
 public class PublishCommand {
     private static final SimpleCommandExceptionType ERROR_FAILED = new SimpleCommandExceptionType(Component.translatable("commands.publish.failed"));
@@ -24,25 +_,43 @@
         dispatcher.register(
             Commands.literal("publish")
                 .requires(Commands.hasPermission(Commands.LEVEL_OWNERS))
-                .executes(c -> publish(c.getSource(), HttpUtil.getAvailablePort(), false))
+                .executes(c -> publish(c.getSource(), HttpUtil.getAvailablePort(), false, null))
                 .then(
                     Commands.argument("allowCommands", BoolArgumentType.bool())
-                        .executes(c -> publish(c.getSource(), HttpUtil.getAvailablePort(), BoolArgumentType.getBool(c, "allowCommands")))
+                        .executes(c -> publish(c.getSource(), HttpUtil.getAvailablePort(), BoolArgumentType.getBool(c, "allowCommands"), null))
                         .then(
-                            Commands.argument("port", IntegerArgumentType.integer(0, 65535))
-                                .executes(c -> publish(c.getSource(), IntegerArgumentType.getInteger(c, "port"), BoolArgumentType.getBool(c, "allowCommands")))
+                            Commands.argument("gamemode", GameModeArgument.gameMode())
+                                .executes(
+                                    c -> publish(
+                                        c.getSource(),
+                                        HttpUtil.getAvailablePort(),
+                                        BoolArgumentType.getBool(c, "allowCommands"),
+                                        GameModeArgument.getGameMode(c, "gamemode")
+                                    )
+                                )
+                                .then(
+                                    Commands.argument("port", IntegerArgumentType.integer(0, 65535))
+                                        .executes(
+                                            c -> publish(
+                                                c.getSource(),
+                                                IntegerArgumentType.getInteger(c, "port"),
+                                                BoolArgumentType.getBool(c, "allowCommands"),
+                                                GameModeArgument.getGameMode(c, "gamemode")
+                                            )
+                                        )
+                                )
                         )
                 )
         );
     }
 
-    private static int publish(CommandSourceStack source, int port, boolean allowCommands) throws CommandSyntaxException {
+    private static int publish(CommandSourceStack source, int port, boolean allowCommands, @Nullable GameType type) throws CommandSyntaxException {
         MinecraftServer server = source.getServer();
         if (server.isPublished() && server.getPort() > -1) {
             throw ERROR_ALREADY_PUBLISHED_LAN.create(server.getPort());
         }
 
-        if (!server.publishServer(MinecraftServer.MultiplayerScope.LAN, allowCommands, port)) {
+        if (!server.publishServer(MinecraftServer.MultiplayerScope.LAN, type, allowCommands, port)) {
             throw ERROR_FAILED.create();
         }
 
