--- a/net/minecraft/server/commands/RecipeCommand.java
+++ b/net/minecraft/server/commands/RecipeCommand.java
@@ -5,7 +_,6 @@
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.function.Predicate;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.commands.arguments.EntityArgument;
@@ -13,13 +_,11 @@
 import net.minecraft.core.registries.Registries;
 import net.minecraft.network.chat.Component;
 import net.minecraft.server.level.ServerPlayer;
-import net.minecraft.world.item.crafting.Recipe;
 import net.minecraft.world.item.crafting.RecipeHolder;
 
 public class RecipeCommand {
     private static final SimpleCommandExceptionType ERROR_GIVE_FAILED = new SimpleCommandExceptionType(Component.translatable("commands.recipe.give.failed"));
     private static final SimpleCommandExceptionType ERROR_TAKE_FAILED = new SimpleCommandExceptionType(Component.translatable("commands.recipe.take.failed"));
-    private static final Predicate<Recipe<?>> LEARNABLE_RECIPE = r -> !r.isSpecial();
     private static final CommandResponseTracker.Messages<ServerPlayer> RESPONSE_GIVE = CommandResponseTracker.messages(
         ERROR_GIVE_FAILED,
         (player, totalValue) -> Component.translatable("commands.recipe.give.success.single", totalValue, player.getDisplayName()),
@@ -40,7 +_,7 @@
                         .then(
                             Commands.argument("targets", EntityArgument.players())
                                 .then(
-                                    Commands.argument("recipe", ResourceKeyArgument.key(Registries.RECIPE, LEARNABLE_RECIPE))
+                                    Commands.argument("recipe", ResourceKeyArgument.key(Registries.RECIPE))
                                         .executes(
                                             c -> giveRecipes(
                                                 c.getSource(),
@@ -55,7 +_,7 @@
                                             c -> giveRecipes(
                                                 c.getSource(),
                                                 EntityArgument.getPlayers(c, "targets"),
-                                                c.getSource().getServer().getRecipeManager().getLearnableRecipes()
+                                                c.getSource().getServer().getRecipeManager().getRecipes()
                                             )
                                         )
                                 )
@@ -66,7 +_,7 @@
                         .then(
                             Commands.argument("targets", EntityArgument.players())
                                 .then(
-                                    Commands.argument("recipe", ResourceKeyArgument.key(Registries.RECIPE, LEARNABLE_RECIPE))
+                                    Commands.argument("recipe", ResourceKeyArgument.key(Registries.RECIPE))
                                         .executes(
                                             c -> takeRecipes(
                                                 c.getSource(),
@@ -81,7 +_,7 @@
                                             c -> takeRecipes(
                                                 c.getSource(),
                                                 EntityArgument.getPlayers(c, "targets"),
-                                                c.getSource().getServer().getRecipeManager().getLearnableRecipes()
+                                                c.getSource().getServer().getRecipeManager().getRecipes()
                                             )
                                         )
                                 )
