--- a/net/minecraft/server/commands/ReloadCommand.java
+++ b/net/minecraft/server/commands/ReloadCommand.java
@@ -2,7 +_,6 @@
 
 import com.google.common.collect.Lists;
 import com.mojang.brigadier.CommandDispatcher;
-import com.mojang.brigadier.context.CommandContext;
 import com.mojang.logging.LogUtils;
 import java.util.Collection;
 import net.minecraft.commands.CommandSourceStack;
@@ -17,6 +_,7 @@
     private static final Logger LOGGER = LogUtils.getLogger();
 
     public static void reloadPacks(Collection<String> p_138236_, CommandSourceStack p_138237_) {
+        p_138237_.getServer().reloadResources$Cause.set(io.papermc.paper.event.server.ServerResourcesReloadedEvent.Cause.COMMAND);
         p_138237_.getServer().reloadResources(p_138236_).exceptionally(p_138234_ -> {
             LOGGER.warn("Failed to execute reload", p_138234_);
             p_138237_.sendFailure(Component.translatable("commands.reload.failure"));
@@ -35,8 +_,19 @@
             }
         }
 
+        net.neoforged.neoforge.resource.ResourcePackLoader.reorderNewlyDiscoveredPacks(collection, p_138225_, p_138223_);
         return collection;
     }
+
+    // CraftBukkit start
+    public static void reload(MinecraftServer minecraftserver) {
+        PackRepository resourcepackrepository = minecraftserver.getPackRepository();
+        WorldData savedata = minecraftserver.getWorldData();
+        Collection<String> collection = resourcepackrepository.getSelectedIds();
+        Collection<String> collection1 = discoverNewPacks(resourcepackrepository, savedata, collection);
+        minecraftserver.reloadResourcesPaper(collection1, io.papermc.paper.event.server.ServerResourcesReloadedEvent.Cause.PLUGIN); // Paper - Add ServerResourcesReloadedEvent);
+    }
+    // CraftBukkit end
 
     public static void register(CommandDispatcher<CommandSourceStack> p_138227_) {
         p_138227_.register(Commands.literal("reload").requires(p_138231_ -> p_138231_.hasPermission(2)).executes(p_288528_ -> {
