--- a/net/minecraft/server/commands/SaveOffCommand.java
+++ b/net/minecraft/server/commands/SaveOffCommand.java
@@ -1,7 +_,6 @@
 package net.minecraft.server.commands;
 
 import com.mojang.brigadier.CommandDispatcher;
-import com.mojang.brigadier.context.CommandContext;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
