--- a/net/minecraft/server/commands/ScheduleCommand.java
+++ b/net/minecraft/server/commands/ScheduleCommand.java
@@ -3,12 +_,10 @@
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.arguments.IntegerArgumentType;
 import com.mojang.brigadier.arguments.StringArgumentType;
-import com.mojang.brigadier.context.CommandContext;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
 import com.mojang.brigadier.suggestion.SuggestionProvider;
-import com.mojang.brigadier.suggestion.SuggestionsBuilder;
 import com.mojang.datafixers.util.Either;
 import com.mojang.datafixers.util.Pair;
 import java.util.Collection;
@@ -31,7 +_,7 @@
         p_304295_ -> Component.translatableEscape("commands.schedule.cleared.failure", p_304295_)
     );
     private static final SuggestionProvider<CommandSourceStack> SUGGEST_SCHEDULE = (p_138424_, p_138425_) -> SharedSuggestionProvider.suggest(
-            p_138424_.getSource().getServer().getWorldData().overworldData().getScheduledEvents().getEventsIds(), p_138425_
+            p_138424_.getSource().getLevel().K.getScheduledEvents().getEventsIds(), p_138425_
         );
 
     public static void register(CommandDispatcher<CommandSourceStack> p_138420_) {
@@ -100,7 +_,7 @@
         } else {
             long i = p_138429_.getLevel().getGameTime() + (long)p_138431_;
             ResourceLocation resourcelocation = p_138430_.getFirst();
-            TimerQueue<MinecraftServer> timerqueue = p_138429_.getServer().getWorldData().overworldData().getScheduledEvents();
+            TimerQueue<MinecraftServer> timerqueue = p_138429_.getLevel().K.overworldData().getScheduledEvents();
             p_138430_.getSecond()
                 .ifLeft(
                     p_305875_ -> {
@@ -133,7 +_,7 @@
     }
 
     private static int remove(CommandSourceStack p_138434_, String p_138435_) throws CommandSyntaxException {
-        int i = p_138434_.getServer().getWorldData().overworldData().getScheduledEvents().remove(p_138435_);
+        int i = p_138434_.getLevel().K.getScheduledEvents().remove(p_138435_);
         if (i == 0) {
             throw ERROR_CANT_REMOVE.create(p_138435_);
         } else {
