--- a/net/minecraft/server/commands/ScheduleCommand.java
+++ b/net/minecraft/server/commands/ScheduleCommand.java
@@ -102,7 +_,7 @@
 
         long tickTime = source.getLevel().getGameTime() + time;
         Identifier callbackId = callback.getFirst();
-        TimerQueue<MinecraftServer> queue = source.getServer().getScheduledEvents();
+        TimerQueue<MinecraftServer> queue = source.getLevel().getScheduledEvents(); // CraftBukkit - SPIGOT-6667: Use world specific function timer
         Optional<CommandFunction<CommandSourceStack>> function = callback.getSecond().left();
         if (function.isPresent()) {
             if (function.get() instanceof MacroFunction) {
