--- a/net/minecraft/server/commands/SetPlayerIdleTimeoutCommand.java
+++ b/net/minecraft/server/commands/SetPlayerIdleTimeoutCommand.java
@@ -2,7 +_,6 @@
 
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.arguments.IntegerArgumentType;
-import com.mojang.brigadier.context.CommandContext;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
 import net.minecraft.network.chat.Component;
