--- a/net/minecraft/server/commands/SetSpawnCommand.java
+++ b/net/minecraft/server/commands/SetSpawnCommand.java
@@ -1,7 +_,6 @@
 package net.minecraft.server.commands;
 
 import com.mojang.brigadier.CommandDispatcher;
-import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import java.util.Collection;
 import java.util.Collections;
 import net.minecraft.commands.CommandSourceStack;
@@ -21,29 +_,6 @@
 import net.minecraft.world.phys.Vec2;
 
 public class SetSpawnCommand {
-    private static final CommandResponseTracker.MessagesWithArg<ServerPlayer, LevelData.RespawnData> RESPONSE_SET = CommandResponseTracker.messages(
-        (player, var1, respawnData) -> Component.translatable(
-            "commands.spawnpoint.success.single",
-            respawnData.pos().getX(),
-            respawnData.pos().getY(),
-            respawnData.pos().getZ(),
-            respawnData.yaw(),
-            respawnData.pitch(),
-            Component.translationArg(respawnData.dimension().identifier()),
-            player.getDisplayName()
-        ),
-        (playerCount, var1, respawnData) -> Component.translatable(
-            "commands.spawnpoint.success.multiple",
-            respawnData.pos().getX(),
-            respawnData.pos().getY(),
-            respawnData.pos().getZ(),
-            respawnData.yaw(),
-            respawnData.pitch(),
-            Component.translationArg(respawnData.dimension().identifier()),
-            playerCount
-        )
-    );
-
     public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
         dispatcher.register(
             Commands.literal("spawnpoint")
@@ -92,20 +_,50 @@
         );
     }
 
-    private static int setSpawn(CommandSourceStack source, Collection<ServerPlayer> targets, BlockPos pos, Coordinates rotation) throws CommandSyntaxException {
-        CommandResponseTracker<ServerPlayer> tracker = CommandResponseTracker.create();
+    private static int setSpawn(CommandSourceStack source, Collection<ServerPlayer> targets, BlockPos pos, Coordinates rotation) {
         ResourceKey<Level> dimension = source.getLevel().dimension();
         Vec2 rotationVector = rotation.getRotation(source);
         float yaw = Mth.wrapDegrees(rotationVector.y);
         float pitch = Mth.clamp(rotationVector.x, -90.0F, 90.0F);
-        LevelData.RespawnData respawnData = LevelData.RespawnData.of(dimension, pos, yaw, pitch);
-        ServerPlayer.RespawnConfig respawnConfig = new ServerPlayer.RespawnConfig(respawnData, true);
 
+        final Collection<ServerPlayer> actualTargets = new java.util.ArrayList<>(); // Paper - Add PlayerSetSpawnEvent
         for (ServerPlayer target : targets) {
-            target.setRespawnPosition(respawnConfig, false);
-            tracker.track(target);
-        }
-
-        return tracker.sendFeedback(source, true, RESPONSE_SET, respawnData);
+            // Paper start - Add PlayerSetSpawnEvent
+            if (target.setRespawnPosition(new ServerPlayer.RespawnConfig(LevelData.RespawnData.of(dimension, pos, yaw, pitch), true), false, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.COMMAND)) {
+                actualTargets.add(target);
+            }
+            // Paper end - Add PlayerSetSpawnEvent
+        }
+        // Paper start - Add PlayerSetSpawnEvent
+        if (actualTargets.isEmpty()) {
+            return 0;
+        }
+        // Paper end - Add PlayerSetSpawnEvent
+
+        String dimensionName = dimension.identifier().toString();
+        if (actualTargets.size() == 1) { // Paper - Add PlayerSetSpawnEvent
+            source.sendSuccess(
+                () -> Component.translatable(
+                    "commands.spawnpoint.success.single",
+                    pos.getX(),
+                    pos.getY(),
+                    pos.getZ(),
+                    yaw,
+                    pitch,
+                    source.getLevel().getDescription(), // Neo: Use dimension translation, if one exists
+                    actualTargets.iterator().next().getDisplayName() // Paper - Add PlayerSetSpawnEvent
+                ),
+                true
+            );
+        } else {
+            source.sendSuccess(
+                () -> Component.translatable(
+                    "commands.spawnpoint.success.multiple", pos.getX(), pos.getY(), pos.getZ(), yaw, pitch, dimensionName, actualTargets.size() // Paper - Add PlayerSetSpawnEvent
+                ),
+                true
+            );
+        }
+
+        return actualTargets.size(); // Paper - Add PlayerSetSpawnEvent
     }
 }
