--- a/net/minecraft/server/commands/StopSoundCommand.java
+++ b/net/minecraft/server/commands/StopSoundCommand.java
@@ -2,7 +_,6 @@
 
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.builder.RequiredArgumentBuilder;
-import com.mojang.brigadier.context.CommandContext;
 import java.util.Collection;
 import javax.annotation.Nullable;
 import net.minecraft.commands.CommandSourceStack;
