--- a/net/minecraft/server/commands/SummonCommand.java
+++ b/net/minecraft/server/commands/SummonCommand.java
@@ -88,6 +_,7 @@
         ServerLevel level = source.getLevel();
         Entity entity = EntityType.loadEntityRecursive(entityTag, level, new EntitySpawnRequest(EntitySpawnReason.COMMAND, false), e -> {
             e.snapTo(pos.x, pos.y, pos.z, e.getYRot(), e.getXRot());
+            e.spawnReason = org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND; // Paper - Entity#getEntitySpawnReason
             return e;
         });
         if (entity == null) {
@@ -98,6 +_,7 @@
             mob.finalizeSpawn(source.getLevel(), source.getLevel().getCurrentDifficultyAt(entity.blockPosition()), EntitySpawnReason.COMMAND, null);
         }
 
+        entity.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND);
         if (!level.tryAddFreshEntityWithPassengers(entity)) {
             throw ERROR_DUPLICATE_UUID.create();
         } else {
