--- a/net/minecraft/server/commands/SummonCommand.java
+++ b/net/minecraft/server/commands/SummonCommand.java
@@ -93,6 +_,7 @@
                     mob.finalizeSpawn(source.getLevel(), source.getLevel().getCurrentDifficultyAt(entity.blockPosition()), EntitySpawnReason.COMMAND, null);
                 }
 
+                entity.spawnReason(org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND);
                 if (!level.tryAddFreshEntityWithPassengers(entity)) {
                     throw ERROR_DUPLICATE_UUID.create();
                 } else {
