--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -19,6 +_,7 @@
 import net.minecraft.core.BlockPos;
 import net.minecraft.network.chat.Component;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.Mth;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
@@ -27,6 +_,10 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec2;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.event.entity.EntityTeleportEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
 import org.jspecify.annotations.Nullable;
 
 public class TeleportCommand {
@@ -244,30 +_,61 @@
         float xRot,
         @Nullable LookAt lookAt
     ) throws CommandSyntaxException {
+        net.neoforged.neoforge.event.entity.EntityTeleportEvent.TeleportCommand event = net.neoforged.neoforge.event.EventHooks.onEntityTeleportCommand(victim, level,x, y, z);
+        if (event.isCanceled()) {
+             return;
+        }
+        x = event.getTargetX();
+        y = event.getTargetY();
+        z = event.getTargetZ();
+
         BlockPos blockPos = BlockPos.containing(x, y, z);
         if (!Level.isInSpawnableBounds(blockPos)) {
             throw INVALID_POSITION.create();
-        }
-
-        double relativeOrAbsoluteX = relatives.contains(Relative.X) ? x - victim.getX() : x;
-        double relativeOrAbsoluteY = relatives.contains(Relative.Y) ? y - victim.getY() : y;
-        double relativeOrAbsoluteZ = relatives.contains(Relative.Z) ? z - victim.getZ() : z;
-        float relativeOrAbsoluteYRot = relatives.contains(Relative.Y_ROT) ? yRot - victim.getYRot() : yRot;
-        float relativeOrAbsoluteXRot = relatives.contains(Relative.X_ROT) ? xRot - victim.getXRot() : xRot;
-        float newYRot = Mth.wrapDegrees(relativeOrAbsoluteYRot);
-        float newXRot = Mth.wrapDegrees(relativeOrAbsoluteXRot);
-        if (victim.teleportTo(level, relativeOrAbsoluteX, relativeOrAbsoluteY, relativeOrAbsoluteZ, relatives, newYRot, newXRot, true)) {
-            if (lookAt != null) {
-                lookAt.perform(source, victim);
-            }
-
-            if (!(victim instanceof LivingEntity living && living.isFallFlying())) {
-                victim.setDeltaMovement(victim.getDeltaMovement().multiply(1.0, 0.0, 1.0));
-                victim.setOnGround(true);
-            }
-
-            if (victim instanceof PathfinderMob mob) {
-                mob.getNavigation().stop();
+        } else {
+            double relativeOrAbsoluteX = relatives.contains(Relative.X) ? x - victim.getX() : x;
+            double relativeOrAbsoluteY = relatives.contains(Relative.Y) ? y - victim.getY() : y;
+            double relativeOrAbsoluteZ = relatives.contains(Relative.Z) ? z - victim.getZ() : z;
+            float relativeOrAbsoluteYRot = relatives.contains(Relative.Y_ROT) ? yRot - victim.getYRot() : yRot;
+            float relativeOrAbsoluteXRot = relatives.contains(Relative.X_ROT) ? xRot - victim.getXRot() : xRot;
+            float newYRot = Mth.wrapDegrees(relativeOrAbsoluteYRot);
+            float newXRot = Mth.wrapDegrees(relativeOrAbsoluteXRot);
+            // CraftBukkit start - Teleport event
+            boolean result;
+            if (victim instanceof ServerPlayer player) {
+                result = player.teleportTo(level, relativeOrAbsoluteX, relativeOrAbsoluteY, relativeOrAbsoluteZ, relatives, newYRot, newXRot, true, PlayerTeleportEvent.TeleportCause.COMMAND);
+            } else {
+                Location to = new Location(level.getWorld(), relativeOrAbsoluteX, relativeOrAbsoluteY, relativeOrAbsoluteZ, newYRot, newXRot);
+                EntityTeleportEvent eventB = new EntityTeleportEvent(victim.getBukkitEntity(), victim.getBukkitEntity().getLocation(), to);
+                level.getCraftServer().getPluginManager().callEvent(eventB);
+                if (eventB.isCancelled()) {
+                    return;
+                }
+
+                relativeOrAbsoluteX = to.getX();
+                relativeOrAbsoluteY = to.getY();
+                relativeOrAbsoluteZ = to.getZ();
+                newYRot = to.getYaw();
+                newXRot = to.getPitch();
+                level = ((CraftWorld) to.getWorld()).getHandle();
+
+                result = victim.teleportTo(level, relativeOrAbsoluteX, relativeOrAbsoluteY, relativeOrAbsoluteZ, relatives, newYRot, newXRot, true);
+            }
+
+            if (result) {
+                // CraftBukkit end
+                if (lookAt != null) {
+                    lookAt.perform(source, victim);
+                }
+
+                if (!(victim instanceof LivingEntity living && living.isFallFlying())) {
+                    victim.setDeltaMovement(victim.getDeltaMovement().multiply(1.0, 0.0, 1.0));
+                    victim.setOnGround(true);
+                }
+
+                if (victim instanceof PathfinderMob mob) {
+                    mob.getNavigation().stop();
+                }
             }
         }
     }
