--- a/net/minecraft/world/level/levelgen/densityfunction/ContextBoundSampler.java
+++ /dev/null
@@ -1,17 +_,0 @@
-package net.minecraft.world.level.levelgen.densityfunction;
-
-import net.minecraft.util.context.ContextKey;
-
-public record ContextBoundSampler(ContextKey<? extends DensitySampler> key, DensitySampler fallbackSampler) implements DensitySampler {
-    @Override
-    public void sampleVolume(SamplerContext context, DensityBuffer outputBuffer, DensityVolume volume) {
-        DensitySampler sampler = context.getFieldOrDefault(this.key, this.fallbackSampler);
-        sampler.sampleVolume(context, outputBuffer, volume);
-    }
-
-    @Override
-    public float sampleValue(SamplerContext context, int blockX, int blockY, int blockZ) {
-        DensitySampler sampler = context.getFieldOrDefault(this.key, this.fallbackSampler);
-        return sampler.sampleValue(context, blockX, blockY, blockZ);
-    }
-}
