--- a/net/minecraft/world/level/levelgen/densityfunction/DensitySamplerSet.java
+++ /dev/null
@@ -1,10 +_,0 @@
-package net.minecraft.world.level.levelgen.densityfunction;
-
-@FunctionalInterface
-public interface DensitySamplerSet {
-    DensitySampler.Bound get(DensityFunction function);
-
-    default float sampleValue(DensityFunction function, int blockX, int blockY, int blockZ) {
-        return this.get(function).sampleValue(blockX, blockY, blockZ);
-    }
-}
