--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerData.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerData.java
@@ -123,7 +_,7 @@
     }
 
     public int countAdditionalPlayers(BlockPos p_312262_) {
-        if (this.detectedPlayers.isEmpty()) {
+        if (false && this.detectedPlayers.isEmpty()) {
             Util.logAndPauseIfInIde("Trial Spawner at " + p_312262_ + " has no detected players");
         }
 
@@ -198,8 +_,8 @@
                 if (p_351984_ instanceof Mob mob) {
                     mob.dropPreservedEquipment();
                 }
-
-                p_351984_.remove(Entity.RemovalReason.DISCARDED);
+                p_351984_.setRemoved$cause(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN);
+                p_351984_.remove(Entity.RemovalReason.DISCARDED); // CraftBukkit - Add bukkit remove cause;
             }
         });
         if (!p_338478_.getOminousConfig().spawnPotentialsDefinition().isEmpty()) {
