--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerConfig.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerConfig.java
@@ -6,8 +_,8 @@
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
-import net.minecraft.core.registries.codec.RegistryCodecs;
 import net.minecraft.nbt.CompoundTag;
+import net.minecraft.resources.RegistryFileCodec;
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.util.random.WeightedList;
 import net.minecraft.world.entity.EntityType;
@@ -54,7 +_,7 @@
             )
             .apply(i, TrialSpawnerConfig::new)
     );
-    public static final Codec<Holder<TrialSpawnerConfig>> CODEC = RegistryCodecs.holder(Registries.TRIAL_SPAWNER_CONFIG, DIRECT_CODEC);
+    public static final Codec<Holder<TrialSpawnerConfig>> CODEC = RegistryFileCodec.create(Registries.TRIAL_SPAWNER_CONFIG, DIRECT_CODEC);
 
     public int calculateTargetTotalMobs(int additionalPlayers) {
         return (int)Math.floor(this.totalMobs + this.totalMobsAddedPerPlayer * additionalPlayers);
