--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerData.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerData.java
@@ -199,7 +_,7 @@
                     mob.dropPreservedEquipment();
                 }
 
-                p_351984_.remove(Entity.RemovalReason.DISCARDED);
+                p_351984_.remove(Entity.RemovalReason.DISCARDED, org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - Add bukkit remove cause;
             }
         });
         if (!p_338478_.getOminousConfig().spawnPotentialsDefinition().isEmpty()) {
