--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
@@ -3,7 +_,6 @@
 import com.google.common.annotations.VisibleForTesting;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
-import com.mojang.serialization.codecs.RecordCodecBuilder.Instance;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import java.util.Optional;
 import java.util.UUID;
@@ -42,6 +_,8 @@
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.CollisionContext;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public final class TrialSpawner implements net.neoforged.neoforge.common.extensions.IOwnedSpawner {
     public static final String NORMAL_CONFIG_TAG_NAME = "normal_config";
@@ -229,15 +_,22 @@
                                 }
 
                                 boolean flag = spawndata.getEntityToSpawn().size() == 1 && spawndata.getEntityToSpawn().contains("id", 8);
-                                if (flag) {
-                                    mob.finalizeSpawn(p_312582_, p_312582_.getCurrentDifficultyAt(mob.blockPosition()), MobSpawnType.TRIAL_SPAWNER, null);
-                                }
+                                // Neo: Patch in FinalizeSpawn for spawners so it may be fired unconditionally, instead of only when vanilla would normally call it.
+                                // The local flag is the conditions under which the spawner will normally call Mob#finalizeSpawn.
+                                net.neoforged.neoforge.event.EventHooks.finalizeMobSpawnSpawner(mob, p_312582_, p_312582_.getCurrentDifficultyAt(mob.blockPosition()), MobSpawnType.TRIAL_SPAWNER, null, this, flag);
 
                                 mob.setPersistenceRequired();
                                 spawndata.getEquipment().ifPresent(mob::equip);
                             }
 
-                            if (!p_312582_.tryAddFreshEntityWithPassengers(entity)) {
+                            entity.spawnedViaMobSpawner = true; // Paper
+                            // CraftBukkit start
+                            entity.spawnReason(CreatureSpawnEvent.SpawnReason.TRIAL_SPAWNER);  // Paper - Entity#getEntitySpawnReason
+                            if (CraftEventFactory.callTrialSpawnerSpawnEvent(entity, p_312518_).isCancelled()) {
+                                return Optional.empty();
+                            }
+                            if (!p_312582_.tryAddFreshEntityWithPassengers(entity, CreatureSpawnEvent.SpawnReason.TRIAL_SPAWNER)) {
+                                // CraftBukkit end
                                 return Optional.empty();
                             } else {
                                 TrialSpawner.FlameParticle trialspawner$flameparticle = this.isOminous
@@ -405,5 +_,14 @@
         TrialSpawnerState getState();
 
         void markUpdated();
+    }
+
+    @Override
+    @org.jetbrains.annotations.Nullable
+    public com.mojang.datafixers.util.Either<net.minecraft.world.level.block.entity.BlockEntity, Entity> getOwner() {
+        if (this.stateAccessor instanceof net.minecraft.world.level.block.entity.TrialSpawnerBlockEntity be) {
+            return com.mojang.datafixers.util.Either.left(be);
+        }
+        return null;
     }
 }
